// igniteui-react-core/esm5/lib/AnchoredCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var AnchoredCategorySeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AnchoredCategorySeriesDescription, _super);
    function AnchoredCategorySeriesDescription() {
        var _this = _super.call(this) || this;
        _this.gl = null;
        _this.gf = null;
        _this.gj = null;
        _this.gk = null;
        _this.gi = null;
        _this.gh = null;
        _this.gd = null;
        _this.f9 = 0;
        _this.f7 = null;
        _this.gb = 0;
        _this.gg = null;
        _this.ge = null;
        return _this;
    }
    AnchoredCategorySeriesDescription.prototype.get_type = function () {
        return "AnchoredCategorySeries";
    };
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "highlightedValueMemberPath", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("HighlightedValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "valueMemberAsLegendLabel", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("ValueMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "valueMemberAsLegendUnit", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("ValueMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "trendLineType", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "markerCollisionAvoidance", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("MarkerCollisionAvoidance");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredCategorySeriesDescription.prototype, "consolidatedItemHitTestBehavior", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("ConsolidatedItemHitTestBehavior");
        },
        enumerable: false,
        configurable: true
    });
    AnchoredCategorySeriesDescription.$t = markType(AnchoredCategorySeriesDescription, 'AnchoredCategorySeriesDescription', CategorySeriesDescription.$);
    return AnchoredCategorySeriesDescription;
}(CategorySeriesDescription));
export { AnchoredCategorySeriesDescription };
