// igniteui-react-core/esm5/lib/AnchoredRadialSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { RadialBaseDescription } from "./RadialBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var AnchoredRadialSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AnchoredRadialSeriesDescription, _super);
    function AnchoredRadialSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.g7 = false;
        _this.hl = null;
        _this.hg = null;
        _this.hj = null;
        _this.hk = null;
        _this.hi = null;
        _this.hh = null;
        _this.hf = null;
        _this.g9 = 0;
        _this.g5 = null;
        _this.hb = 0;
        _this.hc = 0;
        return _this;
    }
    AnchoredRadialSeriesDescription.prototype.get_type = function () {
        return "AnchoredRadialSeries";
    };
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "useCategoryNormalizedValues", {
        get: function () {
            return this.g7;
        },
        set: function (a) {
            this.g7 = a;
            this.g("UseCategoryNormalizedValues");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.hl;
        },
        set: function (a) {
            this.hl = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "highlightedValueMemberPath", {
        get: function () {
            return this.hg;
        },
        set: function (a) {
            this.hg = a;
            this.g("HighlightedValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "valueMemberAsLegendLabel", {
        get: function () {
            return this.hj;
        },
        set: function (a) {
            this.hj = a;
            this.g("ValueMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "valueMemberAsLegendUnit", {
        get: function () {
            return this.hk;
        },
        set: function (a) {
            this.hk = a;
            this.g("ValueMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineType", {
        get: function () {
            return this.hi;
        },
        set: function (a) {
            this.hi = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.hh;
        },
        set: function (a) {
            this.hh = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.hf;
        },
        set: function (a) {
            this.hf = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.g9;
        },
        set: function (a) {
            this.g9 = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.g5;
        },
        set: function (a) {
            this.g5 = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.hb;
        },
        set: function (a) {
            this.hb = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineZIndex", {
        get: function () {
            return this.hc;
        },
        set: function (a) {
            this.hc = a;
            this.g("TrendLineZIndex");
        },
        enumerable: false,
        configurable: true
    });
    AnchoredRadialSeriesDescription.$t = markType(AnchoredRadialSeriesDescription, 'AnchoredRadialSeriesDescription', RadialBaseDescription.$);
    return AnchoredRadialSeriesDescription;
}(RadialBaseDescription));
export { AnchoredRadialSeriesDescription };
