// igniteui-react-core/esm5/lib/BaseDataSourceDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var BaseDataSourceDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BaseDataSourceDescription, _super);
    function BaseDataSourceDescription() {
        var _this = _super.call(this) || this;
        _this.q = false;
        _this.p = false;
        _this.an = 0;
        _this.ao = 0;
        _this.ap = 0;
        _this.n = false;
        _this.h = null;
        _this.i = null;
        _this.j = null;
        _this.aw = null;
        _this.r = false;
        _this.t = false;
        _this.o = false;
        _this.v = false;
        _this.u = false;
        _this.s = false;
        _this.x = false;
        _this.w = false;
        _this.y = false;
        _this.z = false;
        _this.av = null;
        _this.au = null;
        _this.at = null;
        return _this;
    }
    BaseDataSourceDescription.prototype.get_type = function () {
        return "BaseDataSource";
    };
    Object.defineProperty(BaseDataSourceDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "isReadOnly", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("IsReadOnly");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "isBatchingEnabled", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("IsBatchingEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "actualCount", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("ActualCount");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "firstVisibleIndexRequested", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("FirstVisibleIndexRequested");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "lastVisibleIndexRequested", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("LastVisibleIndexRequested");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "deferAutoRefresh", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("DeferAutoRefresh");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "primaryKey", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("PrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "propertiesRequested", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("PropertiesRequested");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "schemaIncludedProperties", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("SchemaIncludedProperties");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "sectionHeaderDisplayMode", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("SectionHeaderDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "isSectionCollapsable", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("IsSectionCollapsable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "isSectionExpandedDefault", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("IsSectionExpandedDefault");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "includeSummaryRowsInSection", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("IncludeSummaryRowsInSection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "isSectionSummaryRowsAtBottom", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("IsSectionSummaryRowsAtBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "isSectionHeaderNormalRow", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("IsSectionHeaderNormalRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "isSectionContentVisible", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("IsSectionContentVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "shouldEmitSectionHeaders", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("ShouldEmitSectionHeaders");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "shouldEmitSectionFooters", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("ShouldEmitSectionFooters");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "shouldEmitShiftedRows", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("ShouldEmitShiftedRows");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "shouldEmitSummaryRows", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("ShouldEmitSummaryRows");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "schemaChangedRef", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("SchemaChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "rowExpansionChangedRef", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("RowExpansionChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseDataSourceDescription.prototype, "rootSummariesChangedRef", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("RootSummariesChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    BaseDataSourceDescription.$t = markType(BaseDataSourceDescription, 'BaseDataSourceDescription', Description.$);
    return BaseDataSourceDescription;
}(Description));
export { BaseDataSourceDescription };
