// igniteui-react-core/esm5/lib/BaseControlTheme.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
export var BaseControlTheme = /*@__PURE__*/ (function (BaseControlTheme) {
    BaseControlTheme[BaseControlTheme["Default"] = 0] = "Default";
    BaseControlTheme[BaseControlTheme["DenaliLight"] = 1] = "DenaliLight";
    BaseControlTheme[BaseControlTheme["MaterialLight"] = 2] = "MaterialLight";
    BaseControlTheme[BaseControlTheme["SlingshotLight"] = 3] = "SlingshotLight";
    BaseControlTheme[BaseControlTheme["SlingshotDark"] = 4] = "SlingshotDark";
    BaseControlTheme[BaseControlTheme["RevealLight"] = 5] = "RevealLight";
    BaseControlTheme[BaseControlTheme["RevealDark"] = 6] = "RevealDark";
    return BaseControlTheme;
})({});
/**
 * @hidden
 */
export var BaseControlTheme_$type = /*@__PURE__*/ markEnum('BaseControlTheme', 'Default,0|DenaliLight,1|MaterialLight,2|SlingshotLight,3|SlingshotDark,4|RevealLight,5|RevealDark,6');
