// igniteui-react-core/esm5/lib/BaseDOMEventProxy_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { EventProxy } from "./EventProxy";
import { Point_$type, runOn, markType, PointUtil, String_$type, Delegate_$type } from "./type";
import { Rect } from "./Rect";
import { MouseEventArgs } from "./MouseEventArgs";
import { Dictionary$2 } from "./Dictionary$2";
import { List$1 } from "./List$1";
/**
 * @hidden
 */
var BaseDOMEventProxy = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BaseDOMEventProxy, _super);
    function BaseDOMEventProxy() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.b4 = null;
        _this.bv = false;
        _this.bx = false;
        _this.b3 = null;
        _this.bo = false;
        _this.b1 = 0;
        _this.b2 = 0;
        _this.bz = 0;
        _this.b0 = 0;
        _this.b5 = 0;
        _this.c3 = null;
        _this.c2 = null;
        _this.bn = false;
        _this.b6 = -1;
        _this.bs = false;
        _this.bp = false;
        _this.bm = false;
        return _this;
    }
    BaseDOMEventProxy.prototype.d = function () {
        var a = new DOMEventProxy(this.b4, this.b3, this.bx);
        a.ab = this.ab;
        a.ap = this.ap;
        return a;
    };
    Object.defineProperty(BaseDOMEventProxy, "by", {
        get: function () {
            return (BaseDOMEventProxy.bt || BaseDOMEventProxy.bu);
        },
        enumerable: false,
        configurable: true
    });
    BaseDOMEventProxy.prototype.bk = function (a) {
        return DOMEventProxy.c4(a);
    };
    BaseDOMEventProxy.prototype.br = function (e_) {
        var pointerEvent_ = e_.originalEvent;
        var a = false;
        if (BaseDOMEventProxy.bu) {
            a = (pointerEvent_.pointerType == 'touch');
        }
        else {
            a = (pointerEvent_.pointerType == pointerEvent_.MSPOINTER_TYPE_TOUCH);
        }
        return a;
    };
    BaseDOMEventProxy.prototype.cu = function (a) {
        if (!this.ae) {
            if (this.bq(a)) {
                if (this.ar == null || this.ar(a)) {
                    a.preventDefault();
                }
            }
        }
    };
    BaseDOMEventProxy.prototype.bq = function (e_) {
        return (e_.originalEvent.cancelable || e_.originalEvent.cancelable == undefined);
    };
    BaseDOMEventProxy.prototype.cq = function () {
        var source_ = this.b4.getNativeElement();
        BaseDOMEventProxy.ct(source_);
    };
    BaseDOMEventProxy.ct = function (source_) {
        if (BaseDOMEventProxy.bu) {
            source_.style.touchAction = 'none';
        }
        else {
            source_.style.msTouchAction = 'none';
        }
        source_.style.msUserSelect = 'none';
        source_.style.webkitTouchCallout = 'none';
        source_.style.webkitUserSelect = 'none';
        source_.style.khtmlUserSelect = 'none';
        source_.style.mozUserSelect = 'none';
        source_.style.userSelect = 'none';
    };
    BaseDOMEventProxy.cn = function (source_) {
        if (BaseDOMEventProxy.bu) {
            source_.style.touchAction = 'auto';
        }
        else {
            source_.style.msTouchAction = 'auto';
        }
        source_.style.msUserSelect = 'auto';
        source_.style.webkitTouchCallout = 'none';
        source_.style.webkitUserSelect = 'auto';
        source_.style.khtmlUserSelect = 'auto';
        source_.style.mozUserSelect = 'auto';
        source_.style.userSelect = 'auto';
    };
    BaseDOMEventProxy.prototype.co = function () {
        var source_ = this.b4.getNativeElement();
        BaseDOMEventProxy.cn(source_);
    };
    BaseDOMEventProxy.prototype.aw = function () {
        _super.prototype.aw.call(this);
        if (!this.ae) {
            this.cq();
        }
        else {
            this.co();
        }
    };
    BaseDOMEventProxy.prototype.cb = function () {
        var ver_ = -1;
        var matchIE_ = /Trident\/([\d.]+)/;
        if ((matchIE_.exec(navigator.userAgent) != null)) {
            ver_ = (parseFloat(RegExp.$1));
        }
        return ver_;
    };
    BaseDOMEventProxy.prototype.ca = function () {
        var ver_ = -1;
        var matchIE_ = /Edge\/([\d.]+)/;
        if ((matchIE_.exec(navigator.userAgent) != null)) {
            ver_ = (parseFloat(RegExp.$1));
        }
        return ver_;
    };
    BaseDOMEventProxy.cs = function (e_) {
        var a = BaseDOMEventProxy.cc >= 6;
        var oe_ = e_.originalEvent;
        if ((((typeof e_.pageX == 'undefined')) || a) && oe_.clientX != null) {
            var od_ = void 0;
            if ((e_.target !== undefined)) {
                od_ = e_.target.ownerDocument;
            }
            else {
                od_ = e_.originalEvent.target.ownerDocument;
            }
            var ed_ = od_ ? od_ : document;
            var doc_ = ed_.documentElement;
            var body_ = ed_.body;
            var clientX_ = oe_.clientX;
            var clientY_ = oe_.clientY;
            var scrollLeft_ = doc_ && doc_.scrollLeft || body_ && body_.scrollLeft || 0;
            var scrollTop_ = doc_ && doc_.scrollTop || body_ && body_.scrollTop || 0;
            var clientLeft_ = doc_ && doc_.clientLeft || body_ && body_.clientLeft || 0;
            var clientTop_ = doc_ && doc_.clientTop || body_ && body_.clientTop || 0;
            e_.pageX = clientX_ + (scrollLeft_ - clientLeft_);
            e_.pageY = clientY_ + (scrollTop_ - clientTop_);
        }
        if ((!e_.pageX)) {
            e_.pageX = e_.originalEvent.pageX;
        }
        if ((!e_.pageY)) {
            e_.pageY = e_.originalEvent.pageY;
        }
        if (((typeof e_.button == 'undefined')) && ((typeof oe_.button != 'undefined'))) {
            e_.button = oe_.button;
        }
    };
    BaseDOMEventProxy.c4 = function (a) {
        var b = BaseDOMEventProxy.cc >= 6;
        var source_ = a.getNativeElement();
        var d_ = source_.ownerDocument;
        var doc_ = d_ ? d_.documentElement : null;
        var z_ = doc_ ? doc_.msContentZoomFactor : null;
        var body_ = doc_.body;
        if ((z_ && z_ > 1) || b) {
            var rect_ = source_.getBoundingClientRect();
            var x_ = (rect_.left);
            var y_ = (rect_.top);
            var scrollLeft_ = (doc_ && doc_.scrollLeft || body_ && body_.scrollLeft || 0);
            var scrollTop_ = (doc_ && doc_.scrollTop || body_ && body_.scrollTop || 0);
            var clientLeft_ = (doc_ && doc_.clientLeft || body_ && body_.clientLeft || 0);
            var clientTop_ = (doc_ && doc_.clientTop || body_ && body_.clientTop || 0);
            var c = x_ + scrollLeft_ - clientLeft_;
            var d = y_ + scrollTop_ - clientTop_;
            return new Rect(0, c, d, 0, 0);
        }
        else {
            var e = a.getOffset();
            return new Rect(0, e.left, e.top, 0, 0);
        }
    };
    Object.defineProperty(BaseDOMEventProxy.prototype, "b8", {
        get: function () {
            return this.b5;
        },
        set: function (a) {
            this.b5 = a;
        },
        enumerable: false,
        configurable: true
    });
    BaseDOMEventProxy.prototype.cx = function (a) {
        this.as = a;
        BaseDOMEventProxy.cs(a);
        var b = BaseDOMEventProxy.c4(this.b4);
        var c = 0;
        var d = 0;
        if (Rect.l_op_Inequality(this.bl, null)) {
            c = this.bl.left;
            d = this.bl.top;
        }
        this.b1 = a.pageX - b.left;
        this.b2 = a.pageY - b.top;
        this.bz = this.b1 - c;
        this.b0 = this.b2 - d;
        this.c3 = { $type: Point_$type, x: this.b1, y: this.b2 };
        this.c2 = { $type: Point_$type, x: this.bz, y: this.b0 };
    };
    BaseDOMEventProxy.prototype.cy = function (e_) {
        this.as = e_;
        BaseDOMEventProxy.cs(e_);
        if (BaseDOMEventProxy.by) {
            var a = e_.pageX;
            var b = e_.pageY;
            var c = BaseDOMEventProxy.c4(this.b4);
            this.b1 = a - c.left;
            this.b2 = b - c.top;
        }
        else {
            this.b5 = 0;
            if (!e_.originalEvent.targetTouches || e_.originalEvent.targetTouches.length < 1) {
                return;
            }
            ;
            this.b5 = (e_.originalEvent.targetTouches.length);
            var d = (e_.originalEvent.targetTouches[0].pageX);
            var e = (e_.originalEvent.targetTouches[0].pageY);
            var f = BaseDOMEventProxy.c4(this.b4);
            this.b1 = d - f.left;
            this.b2 = e - f.top;
        }
        var g = 0;
        var h = 0;
        if (Rect.l_op_Inequality(this.bl, null)) {
            g = this.bl.left;
            h = this.bl.top;
        }
        this.bz = this.b1 - g;
        this.b0 = this.b2 - h;
        this.c3 = { $type: Point_$type, x: this.b1, y: this.b2 };
        this.c2 = { $type: Point_$type, x: this.bz, y: this.b0 };
    };
    BaseDOMEventProxy.prototype.cm = function (e_) {
        this.cw(e_);
        if (this.ad(this.c3, false, true)) {
            var deltaY_ = 0;
            var ev_ = e_;
            deltaY_ = (ev_.originalEvent ? ev_.originalEvent.deltaY : ev_.deltaY);
            var deltaX_ = 0;
            deltaX_ = (ev_.originalEvent ? ev_.originalEvent.deltaX : ev_.deltaX);
            var a = (ev_.originalEvent ? (ev_.originalEvent.deltaMode == 1) : (ev_.deltaMode == 1));
            if (a) {
                deltaY_ = deltaY_ * 40;
                deltaX_ = deltaX_ * 40;
            }
            var b = this.aj(this.c2, deltaY_);
            var c = this.ak(this.c2, deltaX_);
            if (b || c) {
                this.cu(e_);
            }
        }
    };
    BaseDOMEventProxy.prototype.ck = function (e_) {
        this.cw(e_);
        this.cx(e_);
        if (this.ad(this.c3, false, true)) {
            var ev_ = e_;
            var delta_ = 0;
            if (ev_.wheelDelta) {
                delta_ = ev_.wheelDelta / 120;
            }
            ;
            if (ev_.originalEvent && ev_.originalEvent.wheelDelta) {
                delta_ = ev_.originalEvent.wheelDelta / 120;
            }
            ;
            if (ev_.detail) {
                delta_ = -ev_.detail / 3;
            }
            ;
            if (ev_.originalEvent && ev_.originalEvent.detail) {
                delta_ = -ev_.originalEvent.detail / 3;
            }
            ;
            delta_ = delta_ / 10;
            var a = this.aj(this.c2, delta_);
            if (a) {
                this.cu(e_);
            }
        }
    };
    BaseDOMEventProxy.prototype.cr = function () {
        if (this.bn) {
            if (this.b6 == -1) {
                this.b6 = this.b3.setTimeout(runOn(this, this.cp), 500);
            }
            else {
                window.clearTimeout(this.b6);
                this.b6 = -1;
                this.b6 = this.b3.setTimeout(runOn(this, this.cp), 500);
            }
        }
    };
    BaseDOMEventProxy.prototype.cd = function () {
        this.bn = true;
        if (this.b6 != -1) {
            window.clearTimeout(this.b6);
            this.b6 = -1;
        }
    };
    BaseDOMEventProxy.prototype.cp = function () {
        this.bn = false;
    };
    Object.defineProperty(BaseDOMEventProxy.prototype, "bw", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
        },
        enumerable: false,
        configurable: true
    });
    BaseDOMEventProxy.prototype.ci = function (a) {
        if (this.bp || this.bn) {
            return;
        }
        this.cx(a);
        if (this.bs) {
            this.bs = false;
            this.a9(this.c2);
        }
    };
    BaseDOMEventProxy.prototype.cj = function (a) {
        if (this.bp || this.bn) {
            return;
        }
        this.cx(a);
        this.cw(a);
        var b = new MouseEventArgs();
        b.position = this.c3;
        if (this.ad(this.c3, false, false)) {
            if (!this.bs) {
                this.bs = true;
                this.a8(this.c3);
            }
            this.ba(this.c3, true, false);
            this.ay(this.c2, false);
        }
        else if (this.bs) {
            this.ci(a);
        }
    };
    BaseDOMEventProxy.prototype.cv = function (a) {
        this.b4.focus(this.bv);
    };
    BaseDOMEventProxy.prototype.ch = function (a) {
        if (this.bp || this.bn) {
            return;
        }
        this.cv(a);
        this.cx(a);
        this.cw(a);
        if (this.ad(this.c3, false, false)) {
            this.bm = true;
            this.a7(this.c3);
            this.az(this.c2, false);
            this.cu(a);
        }
    };
    BaseDOMEventProxy.prototype.cz = function (a) {
        if (this.bp || this.bn) {
            return;
        }
        if (this.bm) {
            this.cl(a);
        }
    };
    BaseDOMEventProxy.prototype.cl = function (a) {
        if (this.bp || this.bn) {
            return;
        }
        this.cw(a);
        if (PointUtil.equals(this.c3, null) || PointUtil.equals(this.c2, null)) {
            this.cx(a);
        }
        var b = this.ad(this.c3, false, false) || this.bm || this.bs;
        this.bm = false;
        this.bb(this.c3);
        this.ax(this.c2, false, false);
        if (b) {
            this.cu(a);
        }
    };
    BaseDOMEventProxy.prototype.ce = function (a) {
        this.cf(a, true);
    };
    BaseDOMEventProxy.prototype.cg = function (a) {
        this.cf(a, false);
    };
    BaseDOMEventProxy.prototype.c1 = function (a) {
        return BaseDOMEventProxy.c0(a);
    };
    BaseDOMEventProxy.c0 = function (a) {
        var b;
        switch (a.which) {
            case 33:
                b = 10;
                break;
            case 34:
                b = 11;
                break;
            case 36:
                b = 13;
                break;
            case 37:
                b = 14;
                break;
            case 38:
                b = 15;
                break;
            case 39:
                b = 16;
                break;
            case 40:
                b = 17;
                break;
            case 9:
                b = 2;
                break;
            case 32:
                b = 9;
                break;
            case 13:
                b = 3;
                break;
            case 27:
                b = 8;
                break;
            case 16:
                b = 4;
                break;
            case 17:
                b = 5;
                break;
            case 18:
                b = 6;
                break;
            case 8:
                b = 1;
                break;
            case 35:
                b = 12;
                break;
            case 45:
                b = 18;
                break;
            case 46:
                b = 19;
                break;
            case 106:
                b = 78;
                break;
            case 107:
                b = 79;
                break;
            case 109:
                b = 80;
                break;
            case 110:
                b = 81;
                break;
            case 111:
                b = 82;
                break;
            case 220:
                b = 85;
                break;
            case 192:
                b = 86;
                break;
            case 191:
                b = 84;
                break;
            case 186:
                b = 83;
                break;
            case 189:
                b = 88;
                break;
            case 187:
                b = 87;
                break;
            default:
                if (112 <= a.which && a.which <= 123) {
                    b = (56 + a.which - 112);
                }
                else if (65 <= a.which && a.which <= 90) {
                    b = (30 + a.which - 65);
                }
                else if (48 <= a.which && a.which <= 57) {
                    b = (20 + a.which - 48);
                }
                else if (96 <= a.which && a.which <= 105) {
                    b = (68 + a.which - 96);
                }
                else {
                    b = 0;
                }
                break;
        }
        return b;
    };
    BaseDOMEventProxy.prototype.cf = function (a, b) {
        var c = this.c1(a);
        var d = c != 0;
        this.cw(a);
        if (d) {
            var e = false;
            if (b) {
                e = this.ah(c);
            }
            else {
                e = this.ai(c);
            }
            if (e) {
                this.cu(a);
            }
        }
    };
    BaseDOMEventProxy.prototype.cw = function (a) {
        this.as = a;
        this.bi = 0;
        if (a.shiftKey) {
            this.bi = this.bi | 4;
        }
        if (a.altKey) {
            this.bi = this.bi | 1;
        }
        if (a.ctrlKey) {
            this.bi = this.bi | 2;
        }
        this.al = a.button == 2;
    };
    BaseDOMEventProxy.prototype.bj = function () {
        var a = BaseDOMEventProxy.c4(this.b4);
        var b = a.left;
        var c = a.top;
        return { $type: Point_$type, x: b, y: c };
    };
    BaseDOMEventProxy.$t = markType(BaseDOMEventProxy, 'BaseDOMEventProxy', EventProxy.$);
    BaseDOMEventProxy.bt = false;
    BaseDOMEventProxy.bu = false;
    BaseDOMEventProxy.cc = 0;
    BaseDOMEventProxy.b9 = 0;
    return BaseDOMEventProxy;
}(EventProxy));
export { BaseDOMEventProxy };
/**
 * @hidden
 */
var DOMEventProxy = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DOMEventProxy, _super);
    function DOMEventProxy(a, b, c) {
        var _this = _super.call(this) || this;
        _this.de = 0;
        _this.dg = "";
        _this.c6 = false;
        _this.df = null;
        _this.c8 = new Dictionary$2(String_$type, List$1.$.specialize(Delegate_$type), 0);
        _this.d3 = null;
        _this.dc = -1;
        _this.d4 = null;
        _this.da = 0;
        _this.db = 0;
        _this.c9 = 5;
        _this.c5 = false;
        _this.d2 = null;
        _this.bx = c;
        _this.b3 = b;
        _this.ap = function (d) { return true; };
        _this.ab = false;
        DOMEventProxy.dd++;
        _this.de = DOMEventProxy.dd;
        _this.dg = ".DOMProxy" + _this.de.toString();
        _this.bi = 0;
        _this.b4 = a;
        BaseDOMEventProxy.bt = false;
        try {
            BaseDOMEventProxy.bt = (window.navigator.msPointerEnabled && window.MSGesture !== undefined);
        }
        catch (d) {
        }
        BaseDOMEventProxy.bu = false;
        try {
            BaseDOMEventProxy.bu = (window.PointerEvent !== undefined);
        }
        catch (e) {
        }
        BaseDOMEventProxy.cc = _this.cb();
        BaseDOMEventProxy.b9 = _this.ca();
        _this.at(_this.b4, "");
        _this.ao = function (f) { return true; };
        return _this;
    }
    DOMEventProxy.prototype.dh = function (a, b, c) {
    };
    DOMEventProxy.prototype.at = function (a, b) {
        var c = a;
        var d = this.dg + b;
        if (!this.b3.supportsDOMEvents()) {
            return;
        }
        var e = new List$1(Delegate_$type, 0);
        this.c8.addItem(d, e);
        this.dh(a, b, e);
        if (!BaseDOMEventProxy.by) {
            e.add(c.listen("mousemove", runOn(this, this.cj)));
            e.add(c.listen("mouseleave", runOn(this, this.ci)));
            e.add(c.listen("mousedown", runOn(this, this.ch)));
            e.add(c.listen("mouseup", runOn(this, this.cl)));
            e.add(this.b3.globalListen("window", "mouseup", runOn(this, this.cz)));
        }
        e.add(c.listen("keydown", runOn(this, this.ce)));
        e.add(c.listen("keyup", runOn(this, this.cg)));
        var source_ = this.b4.getNativeElement();
        if (!this.ae) {
            BaseDOMEventProxy.ct(source_);
        }
        if (BaseDOMEventProxy.by) {
            var f = this.b4.getNativeElement();
            if ((window.MSGesture !== undefined)) {
                var gesture_ = new window.MSGesture();
                var container_ = f;
                gesture_.target = container_;
                this.df = gesture_;
                e.add(c.listen("MSGestureStart", runOn(this, this.dk)));
                e.add(c.listen("MSGestureChange", runOn(this, this.di)));
                e.add(c.listen("MSGestureEnd", runOn(this, this.dj)));
            }
            if (BaseDOMEventProxy.bu) {
                e.add(c.listen("pointerdown", runOn(this, this.dn)));
                e.add(c.listen("pointerup", runOn(this, this.dr)));
                e.add(c.listen("pointercancel", runOn(this, this.dm)));
                e.add(c.listen("pointermove", runOn(this, this.dp)));
                e.add(c.listen("pointerout", runOn(this, this.dq)));
                e.add(c.listen("lostpointercapture", runOn(this, this.dl)));
            }
            else {
                e.add(c.listen("MSPointerDown", runOn(this, this.dn)));
                e.add(c.listen("MSPointerUp", runOn(this, this.dr)));
                e.add(c.listen("MSPointerCancel", runOn(this, this.dm)));
                e.add(c.listen("MSPointerMove", runOn(this, this.dp)));
                e.add(c.listen("MSPointerOut", runOn(this, this.dq)));
                e.add(c.listen("MSLostPointerCapture", runOn(this, this.dl)));
            }
        }
        else {
            e.add(c.listen("gesturestart", runOn(this, this.dk)));
            e.add(c.listen("gesturechange", runOn(this, this.di)));
            e.add(c.listen("gestureend", runOn(this, this.dj)));
            e.add(c.listen("touchstart", runOn(this, this.du)));
            e.add(c.listen("touchmove", runOn(this, this.dt)));
            e.add(c.listen("touchend", runOn(this, this.ds)));
        }
        if (this.bx) {
            e.add(c.listen("wheel", runOn(this, this.cm)));
        }
        else {
            e.add(c.listen("DOMMouseScroll", runOn(this, this.ck)));
            e.add(c.listen("mousewheel", runOn(this, this.ck)));
        }
    };
    DOMEventProxy.prototype.dr = function (e_) {
        var a = this.br(e_);
        if (a) {
            this.b8 = this.b8 - 1;
            if (this.b8 < 0) {
                this.b8 = 0;
            }
        }
        if (this.b8 < 2 && this.c6 && a) {
            this.c6 = false;
            var b = (e_.originalEvent.scale);
            this.a4(this.c2, b);
        }
        else {
            if (a) {
                this.ds(e_);
            }
            else {
                this.cl(e_);
            }
        }
    };
    DOMEventProxy.prototype.dl = function (e_) {
        var a = this.br(e_);
        if (a) {
            this.b8--;
            if (this.b8 < 0) {
                this.b8 = 0;
            }
        }
        if (this.b8 < 2 && this.c6 && a) {
            this.c6 = false;
            var b = (e_.originalEvent.scale);
            this.a4(this.c2, b);
        }
    };
    DOMEventProxy.prototype.dm = function (e_) {
        var a = this.br(e_);
        if (a) {
            this.b8--;
            if (this.b8 < 0) {
                this.b8 = 0;
            }
        }
        if (this.b8 < 2 && this.c6 && a) {
            this.c6 = false;
            var b = (e_.originalEvent.scale);
            this.a4(this.c2, b);
        }
    };
    DOMEventProxy.prototype.dn = function (e_) {
        var a = this.br(e_);
        if (this.df != null && a) {
            this.b8++;
            var mGesture_ = this.df;
            mGesture_.addPointer(e_.originalEvent.pointerId);
        }
        var eventSource_ = this.b4.getNativeElement();
        if (BaseDOMEventProxy.bu) {
            eventSource_.setPointerCapture(e_.originalEvent.pointerId);
        }
        else {
            eventSource_.msSetPointerCapture(e_.originalEvent.pointerId);
        }
        if (this.b8 > 1 && !this.c6 && a) {
            this.c6 = true;
            this.cy(e_);
            var b = (e_.originalEvent.scale);
            this.bd(this.c2, b);
        }
        else {
            if (a) {
                this.du(e_);
            }
            else {
                this.ch(e_);
            }
        }
    };
    DOMEventProxy.prototype.dp = function (e_) {
        var a = this.br(e_);
        if (this.c6) {
            return;
        }
        if (a) {
            this.dt(e_);
        }
        else {
            this.cj(e_);
        }
    };
    DOMEventProxy.prototype.dq = function (e_) {
        var a = this.br(e_);
        if (this.c6) {
            return;
        }
        if (a) {
        }
        else {
            this.ci(e_);
        }
    };
    DOMEventProxy.prototype.dk = function (e_) {
        this.bs = true;
        this.cw(e_);
        this.cy(e_);
        this.dw();
        if (this.ad(this.c3, true, false)) {
            this.cu(e_);
            if (BaseDOMEventProxy.by && this.b8 < 2) {
                return;
            }
            var a = (e_.originalEvent.scale);
            this.bd(this.c2, a);
        }
    };
    DOMEventProxy.prototype.di = function (e_) {
        this.bs = true;
        this.cw(e_);
        this.cy(e_);
        if (this.ad(this.c3, true, false)) {
            this.cu(e_);
            if (BaseDOMEventProxy.by && this.b8 < 2) {
                return;
            }
            var a = (e_.originalEvent.scale);
            this.bc(this.c2, a);
        }
    };
    DOMEventProxy.prototype.dj = function (e_) {
        this.cu(e_);
        var a = (e_.originalEvent.scale);
        this.bs = false;
        this.cw(e_);
        this.cy(e_);
        if (!BaseDOMEventProxy.by || this.c6) {
            this.a4(this.c2, a);
        }
    };
    DOMEventProxy.prototype.du = function (e_) {
        this.bs = true;
        this.cw(e_);
        this.cy(e_);
        this.d3 = this.c2;
        if (this.ad(this.c3, true, false)) {
            if (!this.ab) {
                this.cu(e_);
            }
            else {
                this.bw = true;
            }
            this.ba(this.c3, false, true);
            this.a7(this.c3);
            this.az(this.c2, true);
            this.dz();
        }
    };
    DOMEventProxy.prototype.dz = function () {
        if (this.dc == -1) {
            this.da = this.c3.x;
            this.db = this.c3.y;
            this.dc = this.b3.setTimeout(runOn(this, this.dx), 1500);
        }
    };
    DOMEventProxy.prototype.dw = function () {
        if (Math.abs(this.da - this.c3.x) > this.c9 || Math.abs(this.db - this.c3.y) > this.c9) {
            this.d0();
        }
    };
    DOMEventProxy.prototype.d0 = function () {
        if (this.dc != -1) {
            window.clearTimeout(this.dc);
            this.dc = -1;
        }
    };
    DOMEventProxy.prototype.dx = function () {
        this.dc = -1;
        if (this.bw) {
            this.bw = false;
            this.cd();
        }
        this.a5(this.c2);
    };
    DOMEventProxy.prototype.dt = function (a) {
        this.bs = true;
        this.cw(a);
        this.cy(a);
        if (this.bw) {
            if (this.ad(this.c3, true, false) && this.c7()) {
                this.bw = false;
                this.cd();
            }
        }
        this.dw();
        this.d1();
        if (this.ad(this.c3, true, false) && this.b8 == 1 && !this.bw) {
            this.cu(a);
            if (!this.c5) {
                this.c5 = true;
                this.d2 = this.c2;
                this.a3(this.d2);
            }
            else {
                this.ba(this.c3, true, true);
                this.ay(this.c2, true);
                this.a2(this.c2);
            }
        }
    };
    DOMEventProxy.prototype.c7 = function () {
        var a = { $type: Point_$type, x: this.d3.x - this.c2.x, y: this.d3.y - this.c2.y };
        if (Math.abs(a.x) < 2 && Math.abs(a.y) < 2) {
            return false;
        }
        if (Math.abs(a.x) > Math.abs(a.y)) {
            a.y = 0;
        }
        if (Math.abs(a.y) > Math.abs(a.x)) {
            a.x = 0;
        }
        return this.ap(a);
    };
    DOMEventProxy.prototype.ds = function (a) {
        this.bs = false;
        this.cw(a);
        this.cy(a);
        this.d0();
        if (this.bw) {
            this.bw = false;
            this.cd();
        }
        this.cu(a);
        this.ba(this.c3, false, true);
        this.bb(this.c3);
        if (this.b8 == 0) {
            this.dy(this.c3);
            this.cr();
        }
        if (this.c5 && this.b8 == 0) {
            this.c5 = false;
            this.d2 = null;
            this.a1(this.c2);
            this.cr();
        }
        this.ax(this.c2, true, this.b8 > 0);
    };
    DOMEventProxy.prototype.dy = function (a) {
        if (Rect.l_op_Equality(this.d4, null)) {
            this.d4 = new Rect(0, a.x - 50, a.y - 50, 100, 100);
            this.b3.setTimeout(runOn(this, this.d1), 500);
        }
        else {
            if (a.x >= this.d4.x && a.x <= this.d4.right && a.y >= this.d4.y && a.y <= this.d4.bottom) {
                this.d1();
                this.a0(a);
            }
        }
    };
    DOMEventProxy.prototype.d1 = function () {
        this.d4 = null;
    };
    DOMEventProxy.prototype.dv = function (a) {
        if (this.c8.containsKey(a)) {
            var b = this.c8.item(a);
            this.c8.removeItem(a);
            for (var c = 0; c < b.count; c++) {
                b._inner[c]();
            }
            b.clear();
        }
    };
    DOMEventProxy.prototype.bh = function (a, b) {
        if (a == null) {
            return;
        }
        var c = a;
        var d = this.dg + b;
        this.dv(d);
    };
    DOMEventProxy.prototype.av = function () {
        if (this.b4 == null) {
            return;
        }
        this.dv(this.dg);
        this.b4 = null;
    };
    DOMEventProxy.$t = markType(DOMEventProxy, 'DOMEventProxy', BaseDOMEventProxy.$);
    DOMEventProxy.dd = 0;
    return DOMEventProxy;
}(BaseDOMEventProxy));
export { DOMEventProxy };
