// igniteui-react-core/esm5/lib/BingMapsMapImageryDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicMapImageryDescription } from "./GeographicMapImageryDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var BingMapsMapImageryDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BingMapsMapImageryDescription, _super);
    function BingMapsMapImageryDescription() {
        var _this = _super.call(this) || this;
        _this.ah = false;
        _this.ag = false;
        _this.aq = null;
        _this.al = null;
        _this.ad = null;
        _this.ac = null;
        _this.an = null;
        _this.ak = null;
        _this.ao = null;
        _this.am = null;
        _this.ap = null;
        return _this;
    }
    BingMapsMapImageryDescription.prototype.get_type = function () {
        return "BingMapsMapImagery";
    };
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "isInitialized", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("IsInitialized");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "isDeferredLoad", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("IsDeferredLoad");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "tilePath", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("TilePath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "actualTilePath", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("ActualTilePath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "subDomains", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("SubDomains");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "actualSubDomains", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("ActualSubDomains");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "bingImageryRestUri", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("BingImageryRestUri");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "actualBingImageryRestUri", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("ActualBingImageryRestUri");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "cultureName", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("CultureName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "apiKey", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("ApiKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BingMapsMapImageryDescription.prototype, "imageryStyle", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("ImageryStyle");
        },
        enumerable: false,
        configurable: true
    });
    BingMapsMapImageryDescription.$t = markType(BingMapsMapImageryDescription, 'BingMapsMapImageryDescription', GeographicMapImageryDescription.$);
    return BingMapsMapImageryDescription;
}(GeographicMapImageryDescription));
export { BingMapsMapImageryDescription };
