// igniteui-react-core/esm5/lib/BulletGraphDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var BulletGraphDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BulletGraphDescription, _super);
    function BulletGraphDescription() {
        var _this = _super.call(this) || this;
        _this.cu = null;
        _this.ac = 0;
        _this.cw = null;
        _this.cv = null;
        _this.dc = null;
        _this.cy = null;
        _this.cp = null;
        _this.a6 = 0;
        _this.c2 = null;
        _this.c3 = null;
        _this.c4 = null;
        _this.at = 0;
        _this.k = null;
        _this.i = null;
        _this.j = null;
        _this.al = 0;
        _this.x = 0;
        _this.ak = 0;
        _this.w = 0;
        _this.aw = 0;
        _this.c6 = null;
        _this.a7 = 0;
        _this.da = null;
        _this.ad = 0;
        _this.v = 0;
        _this.ae = 0;
        _this.cl = null;
        _this.cz = null;
        _this.ar = 0;
        _this.as = 0;
        _this.a8 = 0;
        _this.a9 = 0;
        _this.af = 0;
        _this.a2 = 0;
        _this.a3 = 0;
        _this.ah = 0;
        _this.ag = 0;
        _this.ai = 0;
        _this.aj = 0;
        _this.am = 0;
        _this.a4 = 0;
        _this.a1 = 0;
        _this.a5 = 0;
        _this.c8 = null;
        _this.ct = null;
        _this.c9 = null;
        _this.db = null;
        _this.ba = 0;
        _this.ao = 0;
        _this.an = 0;
        _this.ap = 0;
        _this.c1 = null;
        _this.p = false;
        _this.cq = null;
        _this.cr = null;
        _this.ab = 0;
        _this.z = 0;
        _this.aa = 0;
        _this.av = 0;
        _this.au = 0;
        _this.c5 = null;
        _this.ax = 0;
        _this.ay = 0;
        _this.az = 0;
        _this.c7 = null;
        _this.a0 = 0;
        _this.q = false;
        _this.c0 = null;
        _this.h = null;
        _this.ci = 0;
        _this.ch = 0;
        _this.r = false;
        _this.cs = null;
        _this.aq = 0;
        _this.y = 0;
        _this.cx = null;
        _this.co = null;
        _this.cn = null;
        _this.cm = null;
        return _this;
    }
    Object.defineProperty(BulletGraphDescription.prototype, "fontFamily", {
        get: function () {
            return this.cu;
        },
        set: function (a) {
            this.cu = a;
            this.g("FontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "fontSize", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("FontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "fontWeight", {
        get: function () {
            return this.cw;
        },
        set: function (a) {
            this.cw = a;
            this.g("FontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "fontStyle", {
        get: function () {
            return this.cv;
        },
        set: function (a) {
            this.cv = a;
            this.g("FontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "width", {
        get: function () {
            return this.dc;
        },
        set: function (a) {
            this.dc = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "height", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "background", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    BulletGraphDescription.prototype.get_type = function () {
        return "BulletGraph";
    };
    Object.defineProperty(BulletGraphDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "transitionProgress", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("TransitionProgress");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "orientation", {
        get: function () {
            return this.c2;
        },
        set: function (a) {
            this.c2 = a;
            this.g("Orientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "scaleBackgroundBrush", {
        get: function () {
            return this.c3;
        },
        set: function (a) {
            this.c3 = a;
            this.g("ScaleBackgroundBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "scaleBackgroundOutline", {
        get: function () {
            return this.c4;
        },
        set: function (a) {
            this.c4 = a;
            this.g("ScaleBackgroundOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "scaleBackgroundThickness", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("ScaleBackgroundThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "ranges", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Ranges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "rangeBrushes", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("RangeBrushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "rangeOutlines", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("RangeOutlines");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "minimumValue", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("MinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "actualMinimumValue", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("ActualMinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "maximumValue", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("MaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "actualMaximumValue", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("ActualMaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "targetValue", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("TargetValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "targetValueName", {
        get: function () {
            return this.c6;
        },
        set: function (a) {
            this.c6 = a;
            this.g("TargetValueName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "value", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "valueName", {
        get: function () {
            return this.da;
        },
        set: function (a) {
            this.da = a;
            this.g("ValueName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "highlightValue", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("HighlightValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "actualHighlightValueOpacity", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("ActualHighlightValueOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "highlightValueOpacity", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("HighlightValueOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "actualHighlightValueDisplayMode", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            this.cl = a;
            this.g("ActualHighlightValueDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "highlightValueDisplayMode", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("HighlightValueDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "rangeInnerExtent", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("RangeInnerExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "rangeOuterExtent", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("RangeOuterExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "valueInnerExtent", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("ValueInnerExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "valueOuterExtent", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("ValueOuterExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "interval", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("Interval");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "ticksPostInitial", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("TicksPostInitial");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "ticksPreTerminal", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("TicksPreTerminal");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "labelInterval", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("LabelInterval");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "labelExtent", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("LabelExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "labelsPostInitial", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("LabelsPostInitial");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "labelsPreTerminal", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("LabelsPreTerminal");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "minorTickCount", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("MinorTickCount");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "tickStartExtent", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("TickStartExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "tickEndExtent", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("TickEndExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "tickStrokeThickness", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("TickStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "tickBrush", {
        get: function () {
            return this.c8;
        },
        set: function (a) {
            this.c8 = a;
            this.g("TickBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "fontBrush", {
        get: function () {
            return this.ct;
        },
        set: function (a) {
            this.ct = a;
            this.g("FontBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "valueBrush", {
        get: function () {
            return this.c9;
        },
        set: function (a) {
            this.c9 = a;
            this.g("ValueBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "valueOutline", {
        get: function () {
            return this.db;
        },
        set: function (a) {
            this.db = a;
            this.g("ValueOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "valueStrokeThickness", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("ValueStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "minorTickStartExtent", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("MinorTickStartExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "minorTickEndExtent", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("MinorTickEndExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "minorTickStrokeThickness", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("MinorTickStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "minorTickBrush", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.g("MinorTickBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "isScaleInverted", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("IsScaleInverted");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "backingBrush", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("BackingBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "backingOutline", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("BackingOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "backingStrokeThickness", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("BackingStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "backingInnerExtent", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("BackingInnerExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "backingOuterExtent", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("BackingOuterExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "scaleStartExtent", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("ScaleStartExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "scaleEndExtent", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("ScaleEndExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "targetValueBrush", {
        get: function () {
            return this.c5;
        },
        set: function (a) {
            this.c5 = a;
            this.g("TargetValueBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "targetValueBreadth", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("TargetValueBreadth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "targetValueInnerExtent", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("TargetValueInnerExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "targetValueOuterExtent", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("TargetValueOuterExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "targetValueOutline", {
        get: function () {
            return this.c7;
        },
        set: function (a) {
            this.c7 = a;
            this.g("TargetValueOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "targetValueStrokeThickness", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("TargetValueStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "mergeViewports", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("MergeViewports");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "labelFormat", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("LabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "labelFormatSpecifiers", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("LabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "transitionDuration", {
        get: function () {
            return this.ci;
        },
        set: function (a) {
            this.ci = a;
            this.g("TransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "showToolTipTimeout", {
        get: function () {
            return this.ch;
        },
        set: function (a) {
            this.ch = a;
            this.g("ShowToolTipTimeout");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "showToolTip", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("ShowToolTip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "font", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("Font");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("PixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "actualPixelScalingRatio", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("ActualPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "formatLabelRef", {
        get: function () {
            return this.cx;
        },
        set: function (a) {
            this.cx = a;
            this.g("FormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "alignLabelRef", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.g("AlignLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "actualMinimumValueChangeRef", {
        get: function () {
            return this.cn;
        },
        set: function (a) {
            this.cn = a;
            this.g("ActualMinimumValueChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BulletGraphDescription.prototype, "actualMaximumValueChangeRef", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.g("ActualMaximumValueChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    BulletGraphDescription.$t = markType(BulletGraphDescription, 'BulletGraphDescription', Description.$);
    return BulletGraphDescription;
}(Description));
export { BulletGraphDescription };
