// igniteui-react-core/esm5/lib/ButtonGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ButtonGroupDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ButtonGroupDescription, _super);
    function ButtonGroupDescription() {
        var _this = _super.call(this) || this;
        _this.au = null;
        _this.as = null;
        _this.at = null;
        _this.af = null;
        _this.i = null;
        _this.q = 0;
        _this.ag = null;
        _this.ai = null;
        _this.an = null;
        _this.ak = null;
        _this.ah = null;
        _this.aj = null;
        _this.p = 0;
        _this.ao = null;
        _this.ar = null;
        _this.al = null;
        _this.am = null;
        _this.ap = null;
        _this.aq = null;
        _this.u = 0;
        _this.a8 = null;
        _this.ba = null;
        _this.bf = null;
        _this.bc = null;
        _this.bd = null;
        _this.a9 = null;
        _this.bb = null;
        _this.t = 0;
        _this.be = null;
        _this.bt = null;
        _this.bw = null;
        _this.bu = null;
        _this.bv = null;
        _this.w = 0;
        _this.bh = null;
        _this.bj = null;
        _this.bo = null;
        _this.bl = null;
        _this.bm = null;
        _this.bi = null;
        _this.bk = null;
        _this.v = 0;
        _this.bn = null;
        _this.bp = null;
        _this.bs = null;
        _this.bq = null;
        _this.br = null;
        _this.s = 0;
        _this.av = null;
        _this.ax = null;
        _this.a2 = null;
        _this.az = null;
        _this.a0 = null;
        _this.aw = null;
        _this.ay = null;
        _this.r = 0;
        _this.a1 = null;
        _this.a3 = null;
        _this.a6 = null;
        _this.a4 = null;
        _this.a5 = null;
        _this.bg = null;
        _this.a7 = null;
        _this.l = false;
        _this.m = false;
        _this.h = null;
        _this.bx = null;
        return _this;
    }
    ButtonGroupDescription.prototype.get_type = function () {
        return "ButtonGroup";
    };
    Object.defineProperty(ButtonGroupDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "displayType", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("DisplayType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "baseTheme", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "density", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("Density");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualDensity", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("ActualDensity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectedIndices", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("SelectedIndices");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemCornerRadius", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("ActualItemCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemBackgroundColor", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("ActualItemBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemDisabledBackgroundColor", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("ActualItemDisabledBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemTextColor", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("ActualItemTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemDisabledTextColor", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("ActualItemDisabledTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemBorderColor", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("ActualItemBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemDisabledBorderColor", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("ActualItemDisabledBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemBorderWidth", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("ActualItemBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualSelectedItemBackgroundColor", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("ActualSelectedItemBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualSelectedItemTextColor", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("ActualSelectedItemTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemHoverBackgroundColor", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("ActualItemHoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemHoverTextColor", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("ActualItemHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualSelectedItemHoverBackgroundColor", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("ActualSelectedItemHoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualSelectedItemHoverTextColor", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("ActualSelectedItemHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemCornerRadius", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("ItemCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemBackgroundColor", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("ItemBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemDisabledBackgroundColor", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("ItemDisabledBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemTextColor", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("ItemTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemDisabledTextColor", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("ItemDisabledTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemHoverBackgroundColor", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("ItemHoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemBorderColor", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("ItemBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemDisabledBorderColor", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("ItemDisabledBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemBorderWidth", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("ItemBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemHoverTextColor", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.g("ItemHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectedItemBackgroundColor", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("SelectedItemBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectedItemTextColor", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("SelectedItemTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectedItemHoverBackgroundColor", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("SelectedItemHoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectedItemHoverTextColor", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("SelectedItemHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemCornerRadius", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("OutlinedItemCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemBackgroundColor", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.g("OutlinedItemBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemDisabledBackgroundColor", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("OutlinedItemDisabledBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemTextColor", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            this.bo = a;
            this.g("OutlinedItemTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemDisabledTextColor", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.g("OutlinedItemDisabledTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemHoverBackgroundColor", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            this.bm = a;
            this.g("OutlinedItemHoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemBorderColor", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("OutlinedItemBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemDisabledBorderColor", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("OutlinedItemDisabledBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemBorderWidth", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("OutlinedItemBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemHoverTextColor", {
        get: function () {
            return this.bn;
        },
        set: function (a) {
            this.bn = a;
            this.g("OutlinedItemHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedSelectedItemBackgroundColor", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.g("OutlinedSelectedItemBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedSelectedItemTextColor", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("OutlinedSelectedItemTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedSelectedItemHoverBackgroundColor", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            this.bq = a;
            this.g("OutlinedSelectedItemHoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedSelectedItemHoverTextColor", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.g("OutlinedSelectedItemHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemCornerRadius", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("FlatItemCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemBackgroundColor", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("FlatItemBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemDisabledBackgroundColor", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("FlatItemDisabledBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemTextColor", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("FlatItemTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemDisabledTextColor", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("FlatItemDisabledTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemHoverBackgroundColor", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("FlatItemHoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemBorderColor", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("FlatItemBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemDisabledBorderColor", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("FlatItemDisabledBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemBorderWidth", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("FlatItemBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemHoverTextColor", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("FlatItemHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatSelectedItemBackgroundColor", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("FlatSelectedItemBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatSelectedItemTextColor", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("FlatSelectedItemTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatSelectedItemHoverBackgroundColor", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("FlatSelectedItemHoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatSelectedItemHoverTextColor", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("FlatSelectedItemHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "orientation", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.g("Orientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "id", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "disabled", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "isMultiSelect", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("IsMultiSelect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "buttons", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Buttons");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectionChangedRef", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("SelectionChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    ButtonGroupDescription.$t = markType(ButtonGroupDescription, 'ButtonGroupDescription', Description.$);
    return ButtonGroupDescription;
}(Description));
export { ButtonGroupDescription };
