// igniteui-react-core/esm5/lib/CalloutStyleUpdatingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var CalloutStyleUpdatingEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CalloutStyleUpdatingEventArgsDescription, _super);
    function CalloutStyleUpdatingEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.q = null;
        _this.r = null;
        _this.m = null;
        _this.h = null;
        _this.l = null;
        _this.p = null;
        _this.o = null;
        _this.n = null;
        _this.j = 0;
        return _this;
    }
    CalloutStyleUpdatingEventArgsDescription.prototype.get_type = function () {
        return "CalloutStyleUpdatingEventArgs";
    };
    Object.defineProperty(CalloutStyleUpdatingEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutStyleUpdatingEventArgsDescription.prototype, "xValueRef", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("XValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutStyleUpdatingEventArgsDescription.prototype, "yValueRef", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("YValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutStyleUpdatingEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutStyleUpdatingEventArgsDescription.prototype, "series", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Series");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutStyleUpdatingEventArgsDescription.prototype, "background", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutStyleUpdatingEventArgsDescription.prototype, "textColor", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutStyleUpdatingEventArgsDescription.prototype, "outline", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Outline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutStyleUpdatingEventArgsDescription.prototype, "leaderBrush", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("LeaderBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutStyleUpdatingEventArgsDescription.prototype, "strokeThickness", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("StrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    CalloutStyleUpdatingEventArgsDescription.$t = markType(CalloutStyleUpdatingEventArgsDescription, 'CalloutStyleUpdatingEventArgsDescription', Description.$);
    return CalloutStyleUpdatingEventArgsDescription;
}(Description));
export { CalloutStyleUpdatingEventArgsDescription };
