// igniteui-react-core/esm5/lib/CalloutRenderStyleUpdatingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var CalloutRenderStyleUpdatingEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CalloutRenderStyleUpdatingEventArgsDescription, _super);
    function CalloutRenderStyleUpdatingEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.z = null;
        _this.aa = null;
        _this.j = 0;
        _this.k = 0;
        _this.m = 0;
        _this.n = 0;
        _this.t = null;
        _this.v = null;
        _this.h = null;
        _this.u = null;
        _this.y = null;
        _this.x = null;
        _this.w = null;
        _this.l = 0;
        return _this;
    }
    CalloutRenderStyleUpdatingEventArgsDescription.prototype.get_type = function () {
        return "CalloutRenderStyleUpdatingEventArgs";
    };
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "xValueRef", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("XValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "yValueRef", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("YValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "labelPositionX", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("LabelPositionX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "labelPositionY", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("LabelPositionY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "targetPositionX", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("TargetPositionX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "targetPositionY", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("TargetPositionY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "actualPosition", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("ActualPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "series", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Series");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "background", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "textColor", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "outline", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("Outline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "leaderBrush", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("LeaderBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "strokeThickness", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("StrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    CalloutRenderStyleUpdatingEventArgsDescription.$t = markType(CalloutRenderStyleUpdatingEventArgsDescription, 'CalloutRenderStyleUpdatingEventArgsDescription', Description.$);
    return CalloutRenderStyleUpdatingEventArgsDescription;
}(Description));
export { CalloutRenderStyleUpdatingEventArgsDescription };
