// igniteui-react-core/esm5/lib/CalloutAnnotationDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var CalloutAnnotationDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CalloutAnnotationDescription, _super);
    function CalloutAnnotationDescription() {
        var _this = _super.call(this) || this;
        _this.ae = null;
        _this.af = null;
        _this.ac = null;
        _this.z = null;
        _this.w = null;
        _this.h = null;
        _this.y = null;
        _this.x = null;
        _this.ad = null;
        _this.v = null;
        _this.ab = null;
        _this.aa = null;
        _this.o = 0;
        _this.l = 0;
        _this.n = 0;
        _this.m = 0;
        _this.k = 0;
        _this.j = 0;
        return _this;
    }
    CalloutAnnotationDescription.prototype.get_type = function () {
        return "CalloutAnnotation";
    };
    Object.defineProperty(CalloutAnnotationDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "xValueRef", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("XValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "yValueRef", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("YValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "text", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("Text");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "keyRef", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("KeyRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "contentRef", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("ContentRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "series", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Series");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "itemColor", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("ItemColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "formatLabelRef", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("FormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "textColor", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "background", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "outline", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("Outline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "leaderBrush", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("LeaderBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "strokeThickness", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("StrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "backgroundPaddingLeft", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("BackgroundPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "backgroundPaddingTop", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("BackgroundPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "backgroundPaddingRight", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("BackgroundPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "backgroundPaddingBottom", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("BackgroundPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutAnnotationDescription.prototype, "backgroundCornerRadius", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("BackgroundCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    CalloutAnnotationDescription.$t = markType(CalloutAnnotationDescription, 'CalloutAnnotationDescription', Description.$);
    return CalloutAnnotationDescription;
}(Description));
export { CalloutAnnotationDescription };
