// igniteui-react-core/esm5/lib/CalloutContentUpdatingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var CalloutContentUpdatingEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CalloutContentUpdatingEventArgsDescription, _super);
    function CalloutContentUpdatingEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.k = null;
        _this.i = null;
        _this.h = null;
        return _this;
    }
    CalloutContentUpdatingEventArgsDescription.prototype.get_type = function () {
        return "CalloutContentUpdatingEventArgs";
    };
    Object.defineProperty(CalloutContentUpdatingEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutContentUpdatingEventArgsDescription.prototype, "xValueRef", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("XValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutContentUpdatingEventArgsDescription.prototype, "yValueRef", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("YValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutContentUpdatingEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutContentUpdatingEventArgsDescription.prototype, "contentRef", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ContentRef");
        },
        enumerable: false,
        configurable: true
    });
    CalloutContentUpdatingEventArgsDescription.$t = markType(CalloutContentUpdatingEventArgsDescription, 'CalloutContentUpdatingEventArgsDescription', Description.$);
    return CalloutContentUpdatingEventArgsDescription;
}(Description));
export { CalloutContentUpdatingEventArgsDescription };
