// igniteui-react-core/esm5/lib/CalloutLabelUpdatingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var CalloutLabelUpdatingEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CalloutLabelUpdatingEventArgsDescription, _super);
    function CalloutLabelUpdatingEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.m = null;
        _this.n = null;
        _this.j = null;
        _this.h = null;
        _this.l = null;
        _this.k = null;
        return _this;
    }
    CalloutLabelUpdatingEventArgsDescription.prototype.get_type = function () {
        return "CalloutLabelUpdatingEventArgs";
    };
    Object.defineProperty(CalloutLabelUpdatingEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLabelUpdatingEventArgsDescription.prototype, "xValueRef", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("XValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLabelUpdatingEventArgsDescription.prototype, "yValueRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("YValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLabelUpdatingEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLabelUpdatingEventArgsDescription.prototype, "series", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Series");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLabelUpdatingEventArgsDescription.prototype, "seriesName", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("SeriesName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLabelUpdatingEventArgsDescription.prototype, "labelRef", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("LabelRef");
        },
        enumerable: false,
        configurable: true
    });
    CalloutLabelUpdatingEventArgsDescription.$t = markType(CalloutLabelUpdatingEventArgsDescription, 'CalloutLabelUpdatingEventArgsDescription', Description.$);
    return CalloutLabelUpdatingEventArgsDescription;
}(Description));
export { CalloutLabelUpdatingEventArgsDescription };
