// igniteui-react-core/esm5/lib/CalloutLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var CalloutLayerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CalloutLayerDescription, _super);
    function CalloutLayerDescription() {
        var _this = _super.call(this) || this;
        _this.gd = false;
        _this.g2 = 0;
        _this.g0 = 0;
        _this.g3 = 0;
        _this.g1 = 0;
        _this.h7 = null;
        _this.ib = null;
        _this.ic = null;
        _this.h5 = null;
        _this.gf = false;
        _this.h6 = null;
        _this.ga = null;
        _this.h3 = null;
        _this.h8 = null;
        _this.h9 = null;
        _this.h2 = null;
        _this.f9 = null;
        _this.ge = false;
        _this.gj = false;
        _this.gm = false;
        _this.gi = false;
        _this.gk = false;
        _this.gl = false;
        _this.gn = false;
        _this.h4 = null;
        _this.hp = null;
        _this.gz = 0;
        _this.gh = false;
        _this.gg = false;
        _this.hl = 0;
        _this.h1 = null;
        _this.hw = null;
        _this.ht = null;
        _this.hq = null;
        _this.hm = 0;
        _this.hx = null;
        _this.hv = null;
        _this.g5 = 0;
        _this.g7 = 0;
        _this.g6 = 0;
        _this.g4 = 0;
        _this.hr = null;
        _this.g8 = 0;
        _this.g9 = 0;
        _this.ia = null;
        _this.h0 = null;
        _this.hy = null;
        _this.hs = null;
        _this.hu = null;
        _this.hz = null;
        return _this;
    }
    CalloutLayerDescription.prototype.get_type = function () {
        return "CalloutLayer";
    };
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutExpandsAxisBufferEnabled", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("CalloutExpandsAxisBufferEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutExpandsAxisBufferMinHeight", {
        get: function () {
            return this.g2;
        },
        set: function (a) {
            this.g2 = a;
            this.g("CalloutExpandsAxisBufferMinHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutExpandsAxisBufferMaxHeight", {
        get: function () {
            return this.g0;
        },
        set: function (a) {
            this.g0 = a;
            this.g("CalloutExpandsAxisBufferMaxHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutExpandsAxisBufferMinWidth", {
        get: function () {
            return this.g3;
        },
        set: function (a) {
            this.g3 = a;
            this.g("CalloutExpandsAxisBufferMinWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutExpandsAxisBufferMaxWidth", {
        get: function () {
            return this.g1;
        },
        set: function (a) {
            this.g1 = a;
            this.g("CalloutExpandsAxisBufferMaxWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "labelMemberPath", {
        get: function () {
            return this.h7;
        },
        set: function (a) {
            this.h7 = a;
            this.g("LabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "xMemberPath", {
        get: function () {
            return this.ib;
        },
        set: function (a) {
            this.ib = a;
            this.g("XMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "yMemberPath", {
        get: function () {
            return this.ic;
        },
        set: function (a) {
            this.ic = a;
            this.g("YMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "keyMemberPath", {
        get: function () {
            return this.h5;
        },
        set: function (a) {
            this.h5 = a;
            this.g("KeyMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "isCalloutOffsettingEnabled", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("IsCalloutOffsettingEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "labelFormat", {
        get: function () {
            return this.h6;
        },
        set: function (a) {
            this.h6 = a;
            this.g("LabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "labelFormatSpecifiers", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("LabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "contentMemberPath", {
        get: function () {
            return this.h3;
        },
        set: function (a) {
            this.h3 = a;
            this.g("ContentMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "targetSeriesName", {
        get: function () {
            return this.h8;
        },
        set: function (a) {
            this.h8 = a;
            this.g("TargetSeriesName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "targetSeriesRef", {
        get: function () {
            return this.h9;
        },
        set: function (a) {
            this.h9 = a;
            this.g("TargetSeriesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "collisionChannel", {
        get: function () {
            return this.h2;
        },
        set: function (a) {
            this.h2 = a;
            this.g("CollisionChannel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "allowedPositions", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("AllowedPositions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "isAutoCalloutBehaviorEnabled", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("IsAutoCalloutBehaviorEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useInterpolatedValueForAutoCalloutLabels", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("UseInterpolatedValueForAutoCalloutLabels");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useSeriesColorForOutline", {
        get: function () {
            return this.gm;
        },
        set: function (a) {
            this.gm = a;
            this.g("UseSeriesColorForOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useAutoContrastingLabelColors", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("UseAutoContrastingLabelColors");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useItemColorForFill", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("UseItemColorForFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useItemColorForOutline", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("UseItemColorForOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useValueForAutoCalloutLabels", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("UseValueForAutoCalloutLabels");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "highlightedValueLabelMode", {
        get: function () {
            return this.h4;
        },
        set: function (a) {
            this.h4 = a;
            this.g("HighlightedValueLabelMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "autoCalloutVisibilityMode", {
        get: function () {
            return this.hp;
        },
        set: function (a) {
            this.hp = a;
            this.g("AutoCalloutVisibilityMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutCornerRadius", {
        get: function () {
            return this.gz;
        },
        set: function (a) {
            this.gz = a;
            this.g("CalloutCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "isCustomCalloutStyleEnabled", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("IsCustomCalloutStyleEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "isCustomCalloutRenderStyleEnabled", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("IsCustomCalloutRenderStyleEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "autoCalloutLabelPrecision", {
        get: function () {
            return this.hl;
        },
        set: function (a) {
            this.hl = a;
            this.g("AutoCalloutLabelPrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutTextColor", {
        get: function () {
            return this.h1;
        },
        set: function (a) {
            this.h1 = a;
            this.g("CalloutTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutLightTextColor", {
        get: function () {
            return this.hw;
        },
        set: function (a) {
            this.hw = a;
            this.g("CalloutLightTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutDarkTextColor", {
        get: function () {
            return this.ht;
        },
        set: function (a) {
            this.ht = a;
            this.g("CalloutDarkTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutBackground", {
        get: function () {
            return this.hq;
        },
        set: function (a) {
            this.hq = a;
            this.g("CalloutBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutInterpolatedValuePrecision", {
        get: function () {
            return this.hm;
        },
        set: function (a) {
            this.hm = a;
            this.g("CalloutInterpolatedValuePrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutOutline", {
        get: function () {
            return this.hx;
        },
        set: function (a) {
            this.hx = a;
            this.g("CalloutOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutLeaderBrush", {
        get: function () {
            return this.hv;
        },
        set: function (a) {
            this.hv = a;
            this.g("CalloutLeaderBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutPaddingLeft", {
        get: function () {
            return this.g5;
        },
        set: function (a) {
            this.g5 = a;
            this.g("CalloutPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutPaddingTop", {
        get: function () {
            return this.g7;
        },
        set: function (a) {
            this.g7 = a;
            this.g("CalloutPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutPaddingRight", {
        get: function () {
            return this.g6;
        },
        set: function (a) {
            this.g6 = a;
            this.g("CalloutPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutPaddingBottom", {
        get: function () {
            return this.g4;
        },
        set: function (a) {
            this.g4 = a;
            this.g("CalloutPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutCollisionMode", {
        get: function () {
            return this.hr;
        },
        set: function (a) {
            this.hr = a;
            this.g("CalloutCollisionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutPositionPadding", {
        get: function () {
            return this.g8;
        },
        set: function (a) {
            this.g8 = a;
            this.g("CalloutPositionPadding");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutStrokeThickness", {
        get: function () {
            return this.g9;
        },
        set: function (a) {
            this.g9 = a;
            this.g("CalloutStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "textStyle", {
        get: function () {
            return this.ia;
        },
        set: function (a) {
            this.ia = a;
            this.g("TextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutStyleUpdatingRef", {
        get: function () {
            return this.h0;
        },
        set: function (a) {
            this.h0 = a;
            this.g("CalloutStyleUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutRenderStyleUpdatingRef", {
        get: function () {
            return this.hy;
        },
        set: function (a) {
            this.hy = a;
            this.g("CalloutRenderStyleUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutContentUpdatingRef", {
        get: function () {
            return this.hs;
        },
        set: function (a) {
            this.hs = a;
            this.g("CalloutContentUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutLabelUpdatingRef", {
        get: function () {
            return this.hu;
        },
        set: function (a) {
            this.hu = a;
            this.g("CalloutLabelUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutSeriesSelectingRef", {
        get: function () {
            return this.hz;
        },
        set: function (a) {
            this.hz = a;
            this.g("CalloutSeriesSelectingRef");
        },
        enumerable: false,
        configurable: true
    });
    CalloutLayerDescription.$t = markType(CalloutLayerDescription, 'CalloutLayerDescription', AnnotationLayerDescription.$);
    return CalloutLayerDescription;
}(AnnotationLayerDescription));
export { CalloutLayerDescription };
