// igniteui-react-core/esm5/lib/CellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var CellInfoDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CellInfoDescription, _super);
    function CellInfoDescription() {
        var _this = _super.call(this) || this;
        _this.eh = null;
        _this.b8 = 0;
        _this.y_1 = false;
        _this.ac = false;
        _this.m = false;
        _this.k = false;
        _this.w = false;
        _this.ad = false;
        _this.n = false;
        _this.x_1 = false;
        _this.h = null;
        _this.s = false;
        _this.z = false;
        _this.aa = false;
        _this.v = false;
        _this.r = false;
        _this.cf = 0;
        _this.ec = null;
        _this.bb = 0;
        _this.bc = 0;
        _this.ck = 0;
        _this.cl = 0;
        _this.a4 = 0;
        _this.a2 = 0;
        _this.az = 0;
        _this.a7 = 0;
        _this.a1 = 0;
        _this.ce = 0;
        _this.cm = 0;
        _this.ab = false;
        _this.a9 = 0;
        _this.ay = 0;
        _this.a6 = 0;
        _this.t = false;
        _this.d3 = null;
        _this.ej = null;
        _this.dr = null;
        _this.ds = null;
        _this.dk = null;
        _this.dl = null;
        _this.dj = null;
        _this.dy = null;
        _this.ed = null;
        _this.d4 = null;
        _this.d5 = null;
        _this.ef = null;
        _this.eg = null;
        _this.ea = null;
        _this.d6 = null;
        _this.dz = null;
        _this.a8 = 0;
        _this.d8 = null;
        _this.eb = null;
        _this.ei = null;
        _this.dq = null;
        _this.dt = null;
        _this.d0 = null;
        _this.a5 = 0;
        _this.d1 = null;
        _this.d2 = null;
        _this.dm = null;
        _this.a0 = 0;
        _this.dn = null;
        _this.dp = null;
        _this.d7 = null;
        _this.ba = 0;
        _this.ch = 0;
        _this.cj = 0;
        _this.ci = 0;
        _this.cg = 0;
        _this.b1 = 0;
        _this.b3 = 0;
        _this.b2 = 0;
        _this.b0 = 0;
        _this.b5 = 0;
        _this.b7 = 0;
        _this.b6 = 0;
        _this.b4 = 0;
        _this.bt = 0;
        _this.bv = 0;
        _this.bu = 0;
        _this.bs = 0;
        _this.cb = 0;
        _this.cd = 0;
        _this.cc = 0;
        _this.ca = 0;
        _this.bx = 0;
        _this.bz = 0;
        _this.by = 0;
        _this.bw = 0;
        _this.ee = null;
        _this.q = false;
        _this.l = false;
        _this.d9 = null;
        _this.dv = null;
        _this.a3 = 0;
        _this.dw = null;
        _this.dx = null;
        _this.p = false;
        _this.o = false;
        _this.du = null;
        _this.u = false;
        _this.b9 = 0;
        return _this;
    }
    CellInfoDescription.prototype.get_type = function () {
        return "CellInfo";
    };
    Object.defineProperty(CellInfoDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "styleKey", {
        get: function () {
            return this.eh;
        },
        set: function (a) {
            this.eh = a;
            this.g("StyleKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "dataRow", {
        get: function () {
            return this.b8;
        },
        set: function (a) {
            this.b8 = a;
            this.g("DataRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isPositionDirty", {
        get: function () {
            return this.y_1;
        },
        set: function (a) {
            this.y_1 = a;
            this.g("IsPositionDirty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isSizeDirty", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("IsSizeDirty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isContentDirty", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("IsContentDirty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isBorderDirty", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("IsBorderDirty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isLayerDirty", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("IsLayerDirty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isStateDirty", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("IsStateDirty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isDataDirty", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("IsDataDirty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isPlaceholdContentNeeded", {
        get: function () {
            return this.x_1;
        },
        set: function (a) {
            this.x_1 = a;
            this.g("IsPlaceholdContentNeeded");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actionManager", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ActionManager");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isHitTestVisible", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("IsHitTestVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isRowPinned", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("IsRowPinned");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isRowSticky", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("IsRowSticky");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isLastStickyRow", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("IsLastStickyRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isFilterRow", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("IsFilterRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "indent", {
        get: function () {
            return this.cf;
        },
        set: function (a) {
            this.cf = a;
            this.g("Indent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "rowItemRef", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            this.ec = a;
            this.g("RowItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "x", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("X");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "y", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("Y");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "snappedX", {
        get: function () {
            return this.ck;
        },
        set: function (a) {
            this.ck = a;
            this.g("SnappedX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "snappedY", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            this.cl = a;
            this.g("SnappedY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "editOpacity", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("EditOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "contentOpacity", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("ContentOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualContentOpacity", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("ActualContentOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "opacity", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("Opacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualOpacity", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("ActualOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "height", {
        get: function () {
            return this.ce;
        },
        set: function (a) {
            this.ce = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "width", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isSelected", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("IsSelected");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "selectedStatus", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("SelectedStatus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "activationStatus", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("ActivationStatus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "hoverStatus", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("HoverStatus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isHoverable", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("IsHoverable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "horizontalAlignment", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.g("HorizontalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "verticalAlignment", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("VerticalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "background", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "border", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("Border");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualBackground", {
        get: function () {
            return this.dk;
        },
        set: function (a) {
            this.dk = a;
            this.g("ActualBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualBorder", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            this.dl = a;
            this.g("ActualBorder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "activationBorder", {
        get: function () {
            return this.dj;
        },
        set: function (a) {
            this.dj = a;
            this.g("ActivationBorder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "errorBorder", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("ErrorBorder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "selectedBackground", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            this.ed = a;
            this.g("SelectedBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "hoverBackground", {
        get: function () {
            return this.d4;
        },
        set: function (a) {
            this.d4 = a;
            this.g("HoverBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "hoverTextColor", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("HoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "sortIndicatorColor", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            this.ef = a;
            this.g("SortIndicatorColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "stickyRowBackground", {
        get: function () {
            return this.eg;
        },
        set: function (a) {
            this.eg = a;
            this.g("StickyRowBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "pinnedRowBackground", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.g("PinnedRowBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "lastStickyRowBackground", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            this.d6 = a;
            this.g("LastStickyRowBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "filterRowBackground", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.g("FilterRowBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "pinnedRowOpacity", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("PinnedRowOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "originalValueRef", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.g("OriginalValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "renderValue", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("RenderValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "textColor", {
        get: function () {
            return this.ei;
        },
        set: function (a) {
            this.ei = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualTextColor", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("ActualTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "deletedTextColor", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("DeletedTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "fontFamily", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.g("FontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "fontSize", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("FontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "fontStyle", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            this.d1 = a;
            this.g("FontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "fontWeight", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            this.d2 = a;
            this.g("FontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualFontFamily", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            this.dm = a;
            this.g("ActualFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualFontSize", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("ActualFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualFontStyle", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.g("ActualFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualFontWeight", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("ActualFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "lineBreakMode", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("LineBreakMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "virtualizationPercentage", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("VirtualizationPercentage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "paddingLeft", {
        get: function () {
            return this.ch;
        },
        set: function (a) {
            this.ch = a;
            this.g("PaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "paddingTop", {
        get: function () {
            return this.cj;
        },
        set: function (a) {
            this.cj = a;
            this.g("PaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "paddingRight", {
        get: function () {
            return this.ci;
        },
        set: function (a) {
            this.ci = a;
            this.g("PaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "paddingBottom", {
        get: function () {
            return this.cg;
        },
        set: function (a) {
            this.cg = a;
            this.g("PaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualPaddingLeft", {
        get: function () {
            return this.b1;
        },
        set: function (a) {
            this.b1 = a;
            this.g("ActualPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualPaddingTop", {
        get: function () {
            return this.b3;
        },
        set: function (a) {
            this.b3 = a;
            this.g("ActualPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualPaddingRight", {
        get: function () {
            return this.b2;
        },
        set: function (a) {
            this.b2 = a;
            this.g("ActualPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualPaddingBottom", {
        get: function () {
            return this.b0;
        },
        set: function (a) {
            this.b0 = a;
            this.g("ActualPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "borderLeftWidth", {
        get: function () {
            return this.b5;
        },
        set: function (a) {
            this.b5 = a;
            this.g("BorderLeftWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "borderTopWidth", {
        get: function () {
            return this.b7;
        },
        set: function (a) {
            this.b7 = a;
            this.g("BorderTopWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "borderRightWidth", {
        get: function () {
            return this.b6;
        },
        set: function (a) {
            this.b6 = a;
            this.g("BorderRightWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "borderBottomWidth", {
        get: function () {
            return this.b4;
        },
        set: function (a) {
            this.b4 = a;
            this.g("BorderBottomWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "activationBorderLeftWidth", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("ActivationBorderLeftWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "activationBorderTopWidth", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("ActivationBorderTopWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "activationBorderRightWidth", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("ActivationBorderRightWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "activationBorderBottomWidth", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("ActivationBorderBottomWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "errorBorderLeftWidth", {
        get: function () {
            return this.cb;
        },
        set: function (a) {
            this.cb = a;
            this.g("ErrorBorderLeftWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "errorBorderTopWidth", {
        get: function () {
            return this.cd;
        },
        set: function (a) {
            this.cd = a;
            this.g("ErrorBorderTopWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "errorBorderRightWidth", {
        get: function () {
            return this.cc;
        },
        set: function (a) {
            this.cc = a;
            this.g("ErrorBorderRightWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "errorBorderBottomWidth", {
        get: function () {
            return this.ca;
        },
        set: function (a) {
            this.ca = a;
            this.g("ErrorBorderBottomWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualBorderLeftWidth", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("ActualBorderLeftWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualBorderTopWidth", {
        get: function () {
            return this.bz;
        },
        set: function (a) {
            this.bz = a;
            this.g("ActualBorderTopWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualBorderRightWidth", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            this.by = a;
            this.g("ActualBorderRightWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "actualBorderBottomWidth", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("ActualBorderBottomWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "sortDirection", {
        get: function () {
            return this.ee;
        },
        set: function (a) {
            this.ee = a;
            this.g("SortDirection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isExpanded", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("IsExpanded");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isCollapsable", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IsCollapsable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "pinned", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("Pinned");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "editFontFamily", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.g("EditFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "editFontSize", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("EditFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "editFontStyle", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.g("EditFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "editFontWeight", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("EditFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isEdited", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("IsEdited");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isDeleted", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("IsDeleted");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "editError", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.g("EditError");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "isInEditMode", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("IsInEditMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellInfoDescription.prototype, "editID", {
        get: function () {
            return this.b9;
        },
        set: function (a) {
            this.b9 = a;
            this.g("EditID");
        },
        enumerable: false,
        configurable: true
    });
    CellInfoDescription.$t = markType(CellInfoDescription, 'CellInfoDescription', Description.$);
    CellInfoDescription.__marshalByValue = true;
    CellInfoDescription.__marshalByValueAlias = "CellModel";
    return CellInfoDescription;
}(Description));
export { CellInfoDescription };
