// igniteui-react-core/esm5/lib/CellRangeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var CellRangeDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CellRangeDescription, _super);
    function CellRangeDescription() {
        var _this = _super.call(this) || this;
        _this.k = 0;
        _this.i = 0;
        _this.l = 0;
        _this.j = 0;
        return _this;
    }
    CellRangeDescription.prototype.get_type = function () {
        return "CellRange";
    };
    Object.defineProperty(CellRangeDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellRangeDescription.prototype, "startColumn", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("StartColumn");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellRangeDescription.prototype, "endColumn", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("EndColumn");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellRangeDescription.prototype, "startRow", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("StartRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CellRangeDescription.prototype, "endRow", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("EndRow");
        },
        enumerable: false,
        configurable: true
    });
    CellRangeDescription.$t = markType(CellRangeDescription, 'CellRangeDescription', Description.$);
    CellRangeDescription.__marshalByValue = true;
    CellRangeDescription.__marshalByValueAlias = "CellRange";
    return CellRangeDescription;
}(Description));
export { CellRangeDescription };
