// igniteui-react-core/esm5/lib/ChartMouseEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ChartMouseEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ChartMouseEventArgsDescription, _super);
    function ChartMouseEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.j = null;
        _this.h = null;
        _this.l = null;
        return _this;
    }
    ChartMouseEventArgsDescription.prototype.get_type = function () {
        return "ChartMouseEventArgs";
    };
    Object.defineProperty(ChartMouseEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ChartMouseEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ChartMouseEventArgsDescription.prototype, "series", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Series");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ChartMouseEventArgsDescription.prototype, "plotAreaPosition", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("PlotAreaPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ChartMouseEventArgsDescription.prototype, "chart", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Chart");
        },
        enumerable: false,
        configurable: true
    });
    ChartMouseEventArgsDescription.$t = markType(ChartMouseEventArgsDescription, 'ChartMouseEventArgsDescription', Description.$);
    return ChartMouseEventArgsDescription;
}(Description));
export { ChartMouseEventArgsDescription };
