// igniteui-react-core/esm5/lib/CheckboxDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var CheckboxDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CheckboxDescription, _super);
    function CheckboxDescription() {
        var _this = _super.call(this) || this;
        _this.aq = null;
        _this.a0 = null;
        _this.am = null;
        _this.as = null;
        _this.az = null;
        _this.al = null;
        _this.a1 = null;
        _this.at = null;
        _this.an = null;
        _this.ak = null;
        _this.w = 0;
        _this.t = 0;
        _this.x = 0;
        _this.y = 0;
        _this.v = 0;
        _this.u = 0;
        _this.aj = null;
        _this.av = null;
        _this.au = null;
        _this.aw = null;
        _this.ay = null;
        _this.ag = 0;
        _this.af = 0;
        _this.ap = null;
        _this.ao = null;
        _this.a2 = null;
        _this.l = false;
        _this.ax = null;
        _this.j = false;
        _this.h = false;
        _this.m = false;
        _this.i = false;
        _this.k = false;
        _this.ar = null;
        return _this;
    }
    CheckboxDescription.prototype.get_type = function () {
        return "Checkbox";
    };
    Object.defineProperty(CheckboxDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "baseTheme", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "uncheckedBackgroundColor", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("UncheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "actualUncheckedBackgroundColor", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("ActualUncheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "checkedBackgroundColor", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("CheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "tickColor", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("TickColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "actualTickColor", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("ActualTickColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "uncheckedBorderColor", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("UncheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "checkedBorderColor", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("CheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "actualUncheckedBorderColor", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("ActualUncheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "actualCheckedBorderColor", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("ActualCheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "borderWidth", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("BorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "actualBorderWidth", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("ActualBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "cornerRadius", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("CornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "tickStrokeWidth", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("TickStrokeWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "actualTickStrokeWidth", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("ActualTickStrokeWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "actualCornerRadius", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("ActualCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "actualCheckedBackgroundColor", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("ActualCheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "inputId", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("InputId");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "id", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "labelId", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("LabelId");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "name", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "tabIndex", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("TabIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "required", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("Required");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "ariaLabelledBy", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("AriaLabelledBy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "ariaLabel", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("AriaLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "valueRef", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("ValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "focused", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Focused");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "labelPosition", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("LabelPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "disableRipple", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("DisableRipple");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "checked", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Checked");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "indeterminate", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Indeterminate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "disabled", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "disableTransitions", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("DisableTransitions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CheckboxDescription.prototype, "changeRef", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    CheckboxDescription.$t = markType(CheckboxDescription, 'CheckboxDescription', Description.$);
    return CheckboxDescription;
}(Description));
export { CheckboxDescription };
