// igniteui-react-core/esm5/lib/CodeGeneratingComponentRendererModuleEmitter.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, typeCast, markType } from "./type";
import { CodeGeneratingCodeWriter } from "./CodeGeneratingCodeWriter";
import { CodeGeneratingImportManager } from "./CodeGeneratingImportManager";
import { JsonDictionaryArray } from "./JsonDictionaryArray";
import { JsonDictionaryValue } from "./JsonDictionaryValue";
import { stringStartsWith, stringReplace } from "./string";
/**
 * @hidden
 */
var CodeGeneratingComponentRendererModuleEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CodeGeneratingComponentRendererModuleEmitter, _super);
    function CodeGeneratingComponentRendererModuleEmitter(a, b) {
        var _this = _super.call(this) || this;
        _this.k = null;
        _this.i = null;
        _this.b = null;
        _this.f = new CodeGeneratingImportManager();
        _this.e = new CodeGeneratingImportManager();
        _this.a = null;
        _this.k = a;
        _this.i = b;
        _this.b = new CodeGeneratingCodeWriter(_this.k);
        _this.a = new CodeGeneratingCodeWriter(_this.k);
        return _this;
    }
    Object.defineProperty(CodeGeneratingComponentRendererModuleEmitter.prototype, "j", {
        get: function () {
            return this.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererModuleEmitter.prototype, "d", {
        get: function () {
            return this.b;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererModuleEmitter.prototype, "h", {
        get: function () {
            return this.f;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererModuleEmitter.prototype, "g", {
        get: function () {
            return this.e;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererModuleEmitter.prototype, "c", {
        get: function () {
            return this.a;
        },
        enumerable: false,
        configurable: true
    });
    CodeGeneratingComponentRendererModuleEmitter.prototype.n = function (a, b) {
        var _this = this;
        this.p();
        if (typeCast(JsonDictionaryArray.$, a) !== null) {
            var c = a;
            var _loop_1 = function (d) {
                var e = c.items[d];
                if (typeCast(JsonDictionaryValue.$, e) !== null) {
                    var f = e;
                    if (f != null) {
                        var g_1 = f.value.toString();
                        var h_1;
                        g_1 = ((function () { var i = _this.l(g_1, b, h_1); h_1 = i.p2; return i.ret; })());
                        this_1.m(b, g_1, h_1);
                    }
                }
            };
            var this_1 = this;
            for (var d = 0; d < c.items.length; d++) {
                _loop_1(d);
            }
        }
        this.o();
        this.i.addImportResult("modulesImports", this.h);
        this.i.addContainerResult("modulesRegister", this.d.toString());
        this.i.addImportResult("descriptionImports", this.g);
        this.i.addContainerResult("descriptionRegister", this.c.toString());
    };
    CodeGeneratingComponentRendererModuleEmitter.prototype.l = function (a, b, c) {
        c = false;
        if (a == "withDescriptions") {
            return {
                ret: a,
                p2: c
            };
        }
        var d = "";
        var e = a.split('/');
        if (e.length == 1) {
            d = e[0];
            return {
                ret: d,
                p2: c
            };
        }
        else {
            var f = e[0];
            d = e[1];
            if (stringStartsWith(d, "Web") && b.getMetadata(stringReplace(d, "Module", ""), "__qualifiedNameTS") != null) {
                c = true;
                d = d.substr(3);
                if (stringStartsWith(f, "web")) {
                    this.j.addPackageRef(f);
                }
                else {
                    this.j.addPackageRef("web" + f);
                }
            }
            return {
                ret: f + "/" + d,
                p2: c
            };
        }
    };
    CodeGeneratingComponentRendererModuleEmitter.prototype.o = function () {
    };
    CodeGeneratingComponentRendererModuleEmitter.prototype.m = function (a, b, c) {
    };
    CodeGeneratingComponentRendererModuleEmitter.prototype.p = function () {
    };
    CodeGeneratingComponentRendererModuleEmitter.$t = markType(CodeGeneratingComponentRendererModuleEmitter, 'CodeGeneratingComponentRendererModuleEmitter');
    return CodeGeneratingComponentRendererModuleEmitter;
}(Base));
export { CodeGeneratingComponentRendererModuleEmitter };
