// igniteui-react-core/esm5/lib/CodeGenerationLibraryItemType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
export var CodeGenerationLibraryItemType = /*@__PURE__*/ (function (CodeGenerationLibraryItemType) {
    CodeGenerationLibraryItemType[CodeGenerationLibraryItemType["Data"] = 0] = "Data";
    CodeGenerationLibraryItemType[CodeGenerationLibraryItemType["EventHandler"] = 1] = "EventHandler";
    CodeGenerationLibraryItemType[CodeGenerationLibraryItemType["Template"] = 2] = "Template";
    CodeGenerationLibraryItemType[CodeGenerationLibraryItemType["Unknown"] = 3] = "Unknown";
    return CodeGenerationLibraryItemType;
})({});
/**
 * @hidden
 */
export var CodeGenerationLibraryItemType_$type = /*@__PURE__*/ markEnum('CodeGenerationLibraryItemType', 'Data,0|EventHandler,1|Template,2|Unknown,3');
