// igniteui-react-core/esm5/lib/CodeGeneratingComponentRendererCodeEmitter_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { Base, NotSupportedException, IList_$type, String_$type, typeCast, Array_$type, fromEnum, Boolean_$type, markType, EnumUtil, fromEn, Number_$type, BaseError, getInstanceType, Point_$type, typeGetValue, runOn } from "./type";
import { CodeGenerationTargetPlatforms_$type } from "./CodeGenerationTargetPlatforms";
import { CodeGeneratingRendererResult } from "./CodeGeneratingRendererResult";
import { CodeGeneratingCodeWriter } from "./CodeGeneratingCodeWriter";
import { Stack$1 } from "./Stack$1";
import { CodeGenerationItemBuilderPropertyInfo } from "./CodeGenerationItemBuilderPropertyInfo";
import { List$1 } from "./List$1";
import { TypeDescriptionContext } from "./TypeDescriptionContext";
import { TypeDescriptionMetadata } from "./TypeDescriptionMetadata";
import { Rect } from "./Rect";
import { Size } from "./Size";
import { CodeGenerationSpecialValue } from "./CodeGenerationSpecialValue";
import { ColorUtil } from "./ColorUtil";
import { Brush } from "./Brush";
import { StringBuilder } from "./StringBuilder";
import { CodeGeneratingImportManager } from "./CodeGeneratingImportManager";
import { stringJoin1, stringIsNullOrEmpty, stringReplace, stringStartsWith, stringEndsWith, stringContains } from "./string";
import { HashSet$1 } from "./HashSet$1";
import { arrayGetValue } from "./array";
import { trimEnd } from "./stringExtended";
import { WebImportsHalper } from "./WebImportsHalper";
import { CodeGeneratingComponentRendererHandlerEmitter } from "./CodeGeneratingComponentRendererHandlerEmitter";
import { Dictionary$2 } from "./Dictionary$2";
import { CodeGeneratingRegexHelper } from "./CodeGeneratingRegexHelper";
import { CodeGeneratingComponentRendererDataEmitter } from "./CodeGeneratingComponentRendererDataEmitter";
import { JsonDictionaryValue } from "./JsonDictionaryValue";
import { JsonDictionaryObject } from "./JsonDictionaryObject";
import { JsonDictionaryValueType_$type } from "./JsonDictionaryValueType";
import { JsonDictionaryArray } from "./JsonDictionaryArray";
import { isNaN_ } from "./number";
import { timeSpanFromMilliseconds } from "./timespan";
import { ComponentRenderer } from "./ComponentRenderer";
import { DotNetCodeGeneratingComponentRendererHandlerEmitter } from "./DotNetCodeGeneratingComponentRendererHandlerEmitter";
import { TSCodeGeneratingComponentRendererTemplateEmitter } from "./TSCodeGeneratingComponentRendererTemplateEmitter";
import { DotNetCodeGeneratingComponentRendererTemplateEmitter } from "./DotNetCodeGeneratingComponentRendererTemplateEmitter";
import { TSCodeGeneratingComponentRendererModuleEmitter } from "./TSCodeGeneratingComponentRendererModuleEmitter";
import { DotNetCodeGeneratingComponentRendererModuleEmitter } from "./DotNetCodeGeneratingComponentRendererModuleEmitter";
import { BlazorCodeGeneratingComponentRendererModuleEmitter } from "./BlazorCodeGeneratingComponentRendererModuleEmitter";
import { CodeGenerationTransformer } from "./CodeGenerationTransformer";
import { LinearGradientBrush } from "./LinearGradientBrush";
import { IComponentRendererAdapter_$type } from "./IComponentRendererAdapter";
/**
 * @hidden
 */
var CodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CodeGeneratingComponentRendererCodeEmitter, _super);
    function CodeGeneratingComponentRendererCodeEmitter(a, b) {
        var _this = _super.call(this) || this;
        _this.l = null;
        _this.j = null;
        _this.z = new List$1(CodeGenerationItemBuilderPropertyInfo.$, 0);
        _this.d = null;
        _this.a = null;
        _this.c = null;
        _this.b = null;
        _this.aa = new List$1(String_$type, 0);
        _this.af = new Stack$1(String_$type);
        _this.ad = new Stack$1(String_$type);
        _this.ae = new Stack$1(Boolean_$type);
        _this.a5 = null;
        _this.a4 = null;
        _this.l = a;
        _this.j = b;
        _this.d = new CodeGeneratingCodeWriter(_this.l);
        _this.a = new CodeGeneratingCodeWriter(_this.l);
        _this.c = new CodeGeneratingCodeWriter(_this.l);
        _this.b = new CodeGeneratingCodeWriter(_this.l);
        return _this;
    }
    CodeGeneratingComponentRendererCodeEmitter.prototype.get_al = function () {
        switch (this.j.platform) {
            case 1: return "Igx";
            case 2: return "Igr";
            case 3: return "Igc";
            case 4: return "Ultra";
            case 0: return "Xam";
            case 5: return "Igb";
        }
        return "";
    };
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "al", {
        get: function () {
            return this.get_al();
        },
        enumerable: false,
        configurable: true
    });
    CodeGeneratingComponentRendererCodeEmitter.prototype.get_am = function () {
        switch (this.j.platform) {
            case 1: return "Component";
            case 2: return "";
            case 3: return "Component";
            case 4: return "";
            case 0: return "";
        }
        return "";
    };
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "am", {
        get: function () {
            return this.get_am();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "a6", {
        get: function () {
            switch (this.j.platform) {
                case 1: return "angular";
                case 2: return "react";
                case 3: return "webcomponents";
            }
            return "";
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "m", {
        get: function () {
            return this.l;
        },
        enumerable: false,
        configurable: true
    });
    CodeGeneratingComponentRendererCodeEmitter.prototype.br = function (a) {
        this.j.af(a);
    };
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "k", {
        get: function () {
            return this.j;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "h", {
        get: function () {
            return this.d;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "e", {
        get: function () {
            return this.a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "g", {
        get: function () {
            return this.c;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "f", {
        get: function () {
            return this.b;
        },
        enumerable: false,
        configurable: true
    });
    CodeGeneratingComponentRendererCodeEmitter.prototype.bb = function (a) {
        if (a == null || a == "any") {
            return;
        }
        this.aa.add(a);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.w = function (a, b) {
        if (this.r) {
            return false;
        }
        if (a.hasMustSetInCodePlatforms(b.b.owningType, b.b.propertyName)) {
            var c = a.getMustSetInCodePlatforms(b.b.owningType, b.b.propertyName);
            var _loop_1 = function (e) {
                var d = c[e];
                var f;
                if (((function () { var g = EnumUtil.tryParse$1(CodeGenerationTargetPlatforms_$type, d, true, f); f = g.p2; return g.ret; })())) {
                    if (f == this_1.j.platform) {
                        return { value: true };
                    }
                }
            };
            var this_1 = this;
            for (var e = 0; e < c.length; e++) {
                var state_1 = _loop_1(e);
                if (typeof state_1 === "object")
                    return state_1.value;
            }
        }
        return false;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.q = function (a, b) {
        if (b.b != null && (b.b.knownType == 20 || b.b.knownType == 19 || b.b.knownType == 22 || b.b.knownType == 18)) {
            if (b.a != null) {
                return true;
            }
        }
        if (b.b != null && (b.b.knownType == 8 || b.b.knownType == 9 || b.b.knownType == 10)) {
            if (this.w(a, b)) {
                return true;
            }
        }
        return false;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.t = function (a) {
        if (a.b != null && (a.b.knownType == 19)) {
            if (a.a != null) {
                return true;
            }
        }
        return false;
    };
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "aq", {
        get: function () {
            return this.af.d();
        },
        enumerable: false,
        configurable: true
    });
    CodeGeneratingComponentRendererCodeEmitter.prototype.bc = function (a, b, c) {
        this.af.h(a);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.u = function (a, b) {
        var c = false;
        if (b.getMetadata(a, "__isPlainObject") != null) {
            c = true;
        }
        return c;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bq = function () {
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bk = function (a) {
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bl = function (a) {
        this.af.e();
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bi = function (a, b, c, d) {
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bj = function (a, b, c, d) {
        this.bi(a, b, c, d);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.toString = function () {
        return this.d.toString();
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.a0 = function () {
        return "null";
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.a3 = function (a, b, c) {
        if (b.b == null) {
            return b.c;
        }
        if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
            b.c = this.ag(b);
        }
        switch (b.b.knownType) {
            case 10:
                if (this.w(a, b)) {
                    return this.ao(a, b);
                }
                return this.at(a, b);
            case 23:
                if (this.w(a, b)) {
                    return this.ao(a, b);
                }
                return this.at(a, b);
            case 7: return b.c ? "true" : "false";
            case 4: return this.ai(b);
            case 6:
                {
                    return stringJoin1(String_$type, " ", this.ab(b, c));
                }
            case 9:
                if (this.w(a, b)) {
                    return this.ao(a, b);
                }
                throw new NotSupportedException(1, "shouldn't get here");
            case 5: return this.ak(b);
            case 16:
                {
                    return stringJoin1(String_$type, " ", this.ac(b, c));
                }
            case 20:
                if (b.a != null) {
                    return this.a2(b);
                }
                if (b.c == null) {
                    return this.a0();
                }
                break;
            case 22:
                if (b.a != null) {
                    return this.a2(b);
                }
                if (b.c == null) {
                    return this.a0();
                }
                break;
            case 15: throw new NotSupportedException(1, "need to figure this out");
            case 3: throw new NotSupportedException(1, "need to figure this out");
            case 19: if (b.a != null) {
                return this.ax(b);
            }
            else {
                return this.a0();
            }
            case 8:
                if (b.b.specificExternalType == "string") {
                    return this.aw(b);
                }
                if (b.a != null) {
                    return this.ay(b);
                }
                if (this.w(a, b)) {
                    return this.ao(a, b);
                }
                throw new NotSupportedException(1, "shouldn't get here");
            case 13: throw new NotSupportedException(1, "shouldn't get here");
            case 18: if (b.a != null) {
                return this.az(b);
            }
            else {
                return this.a0();
            }
            case 1: return b.c.toString();
            case 11:
                {
                    var d = b.c;
                    return d.x + ", " + d.y;
                }
            case 14:
                {
                    var e = b.c;
                    return e.left + ", " + e.top + ", " + e.width + ", " + e.height;
                }
            case 12:
                {
                    var f = b.c;
                    return f.width + ", " + f.height;
                }
            case 2: return b.c != null ? b.c.toString() : this.a0();
            case 21: return this.a9(b.c);
            case 17:
                if (b.c == null) {
                    return "null";
                }
                if (typeof b.c === 'string') {
                    return b.c.toString();
                }
                return b.c.toString();
            case 0: throw new NotSupportedException(1, "shouldn't get here");
        }
        return "";
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.ab = function (a, b) {
        if (a.b == null) {
            return ((function () {
                var $ret = new List$1(String_$type, 0);
                $ret.add(a.c);
                return $ret;
            })());
        }
        if (typeCast(CodeGenerationSpecialValue.$, a.c) !== null) {
            a.c = this.ag(a);
        }
        var c = a.c;
        var d = new List$1(String_$type, 0);
        for (var e = 0; e < c.count; e++) {
            d.add(this.ah(c.item(e)));
        }
        return d;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.ac = function (a, b) {
        if (a.b == null) {
            return ((function () {
                var $ret = new List$1(String_$type, 0);
                $ret.add(a.c);
                return $ret;
            })());
        }
        if (typeCast(CodeGenerationSpecialValue.$, a.c) !== null) {
            a.c = this.ag(a);
        }
        var c = a.c;
        var d = new List$1(String_$type, 0);
        for (var e = 0; e < c.count; e++) {
            d.add(this.aj(c.item(e)));
        }
        return d;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.a9 = function (a) {
        if (a == null) {
            return "";
        }
        return a.toString();
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.ao = function (a, b) {
        var c = this.a8(a, b);
        if (b.c != null) {
            var d = b.c;
            if (typeCast(List$1.$.specialize(Base.$), d) !== null) {
                d = d.toArray();
            }
            if (typeCast(CodeGenerationItemBuilder.$, d) !== null) {
                var e = this.i(this.l, this.j);
                var f = this.an(c, b.c, e);
                c = this.ar(a, b.c, f, c);
            }
            else if (typeCast(Array_$type, d) !== null) {
                var g = d;
                var h = new List$1(CodeGenerationItemBuilder.$, 0);
                var i = new List$1(String_$type, 0);
                for (var j = 0; j < g.length; j++) {
                    if (typeCast(CodeGenerationItemBuilder.$, g[j]) !== null) {
                        var k = this.i(this.l, this.j);
                        var l = c;
                        var m = g[j].p();
                        if (m == null) {
                            m = g[j].n();
                            m = this.au(m);
                        }
                        if (m != null) {
                            if (m != l) {
                                l = m;
                            }
                            else {
                                l = c + "_" + j.toString();
                            }
                        }
                        var n = this.an(l, g[j], k);
                        h.add(g[j]);
                        i.add(n);
                    }
                }
                c = this.as(b, a, h, i, c);
            }
        }
        return c;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.an = function (a, b, c) {
        var e_1, _a;
        b.u = a;
        b.v(c, true);
        var d = c.toString();
        try {
            for (var _b = __values(fromEnum(c.aa)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var e = _c.value;
                this.aa.add(e);
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        return d;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.ar = function (a, b, c, d) {
        return d;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.as = function (a, b, c, d, e) {
        return e;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.a8 = function (a, b) {
        if (b.c == null) {
            return this.a0();
        }
        if (typeCast(CodeGenerationItemBuilder.$, b.c) !== null) {
            var c = b.c.p();
            if (c != null) {
                return c;
            }
            var d = b.c.n();
            var e = this.au(d);
            return e;
        }
        else if (typeCast(IList_$type, b.c) !== null) {
            var f = b.c;
            if (f.count > 0 && typeCast(CodeGenerationItemBuilder.$, f.item(0)) !== null) {
                var g = f.item(0);
                var h = g.p();
                if (h != null) {
                    return h;
                }
                var i = g.n();
                var j = this.au(i);
                return j;
            }
            return this.au(b.b.propertyName);
        }
        else if (typeCast(Array_$type, b.c) !== null) {
            var k = b.c;
            if (k.length > 0 && typeCast(CodeGenerationItemBuilder.$, k[0]) !== null) {
                var l = b.c.p();
                if (l != null) {
                    return l;
                }
                var m = b.c.n();
                var n = this.au(m);
                return n;
            }
            return this.au(b.b.propertyName);
        }
        return this.a0();
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.at = function (a, b) {
        return "";
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.ag = function (a) {
        return a.c;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.a2 = function (a) {
        if (!stringIsNullOrEmpty(a.a.e)) {
            if (a.a.a) {
                return this.av(a);
            }
            else {
                return this.a1(a);
            }
        }
        return this.a0();
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.s = function (a) {
        if (a.a != null && !stringIsNullOrEmpty(a.a.e)) {
            if (a.a.a) {
                return true;
            }
        }
        return false;
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.a1 = function (a) {
        return "";
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.av = function (a) {
        return "";
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.ay = function (a) {
        if (!stringIsNullOrEmpty(a.a.e)) {
            return a.a.e;
        }
        return this.a0();
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.aw = function (a) {
        return a.c != null ? a.c.toString() : "";
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.az = function (a) {
        return this.ax(a);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.ax = function (a) {
        if (!stringIsNullOrEmpty(a.a.e)) {
            this.br(a);
            return a.a.e;
        }
        return this.a0();
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.get_x = function () {
        return false;
    };
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "x", {
        get: function () {
            return this.get_x();
        },
        enumerable: false,
        configurable: true
    });
    CodeGeneratingComponentRendererCodeEmitter.prototype.ak = function (a) {
        var b = a.c;
        return ColorUtil.k(b, this.x);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.aj = function (a) {
        var b = a;
        return ColorUtil.k(b, this.x);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.ai = function (a) {
        if (a.c == null) {
            return "";
        }
        var b = a.c;
        return ColorUtil.k(b.color, this.x);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.ah = function (a) {
        if (a == null) {
            return "";
        }
        var b = a;
        return ColorUtil.k(b.color, this.x);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bp = function (a) {
        this.ad.h(a);
        this.ae.h(false);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bo = function (a) {
        var b = this.au(a);
        this.ad.h(b);
        this.ae.h(true);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.au = function (a) {
        var b = stringReplace(stringReplace(a, this.a7, ""), "Component", "");
        b = CodeGeneratingComponentRenderer.e5(b);
        return this.j.v(b);
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bm = function () {
        this.a4 = this.ad.e();
        this.ae.e();
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.get_a7 = function () {
        return "";
    };
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "a7", {
        get: function () {
            return this.get_a7();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "ap", {
        get: function () {
            if (this.ad.f > 0) {
                return this.ad.d();
            }
            var a = stringReplace(stringReplace(this.aq, this.a7, ""), "Component", "");
            return CodeGeneratingComponentRenderer.e5(a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "v", {
        get: function () {
            return this.af.f == 1;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "o", {
        get: function () {
            if (this.ae.f == 0) {
                return false;
            }
            return this.ae.d();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "p", {
        get: function () {
            if (this.ad.f > 0) {
                return true;
            }
            return false;
        },
        enumerable: false,
        configurable: true
    });
    CodeGeneratingComponentRendererCodeEmitter.prototype.get_y = function () {
        return true;
    };
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "y", {
        get: function () {
            return this.get_y();
        },
        enumerable: false,
        configurable: true
    });
    CodeGeneratingComponentRendererCodeEmitter.prototype.get_n = function () {
        return this.r;
    };
    Object.defineProperty(CodeGeneratingComponentRendererCodeEmitter.prototype, "n", {
        get: function () {
            return this.get_n();
        },
        enumerable: false,
        configurable: true
    });
    CodeGeneratingComponentRendererCodeEmitter.prototype.bn = function () {
        this.a4 = this.ad.e();
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.be = function (a) {
        var b = this.f.toString();
        if (b.length > 0) {
            a.l(b);
        }
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bh = function (a) {
        var b = this.g.toString();
        if (b.length > 0) {
            a.l(b);
        }
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bd = function (a) {
        var b = this.e.toString();
        if (b.length > 0) {
            a.l(b);
        }
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bf = function (a, b) {
        if (this.aa.count > 0) {
            this.bg(a, this.aa, b);
        }
    };
    CodeGeneratingComponentRendererCodeEmitter.prototype.bg = function (a, b, c) {
    };
    CodeGeneratingComponentRendererCodeEmitter.$t = markType(CodeGeneratingComponentRendererCodeEmitter, 'CodeGeneratingComponentRendererCodeEmitter');
    return CodeGeneratingComponentRendererCodeEmitter;
}(Base));
export { CodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var BlazorRazorCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BlazorRazorCodeGeneratingComponentRendererCodeEmitter, _super);
    function BlazorRazorCodeGeneratingComponentRendererCodeEmitter(a, b) {
        var _this = _super.call(this, a, b) || this;
        _this.bu = 0;
        _this.bs = new HashSet$1(String_$type, 0);
        _this.bt = new HashSet$1(String_$type, 0);
        return _this;
    }
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.get_r = function () {
        return false;
    };
    Object.defineProperty(BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype, "r", {
        get: function () {
            return this.get_r();
        },
        enumerable: false,
        configurable: true
    });
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.ar = function (a, b, c, d) {
        d = _super.prototype.ar.call(this, a, b, c, d);
        this.f.l("private " + b.n() + " _" + CodeGeneratingComponentRenderer.e5(d) + " = null;");
        this.f.l("public " + b.n() + " " + CodeGeneratingComponentRenderer.e6(d));
        this.f.l("{");
        this.f.h();
        this.f.l("get");
        this.f.l("{");
        this.f.h();
        this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(d) + " == null)");
        this.f.l("{");
        this.f.h();
        this.f.k(c);
        this.f.l("this._" + CodeGeneratingComponentRenderer.e5(d) + " = " + d + ";");
        this.f.f();
        this.f.l("}");
        this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(d) + ";");
        this.f.f();
        this.f.l("}");
        this.f.f();
        this.f.l("}");
        return CodeGeneratingComponentRenderer.e6(d);
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.as = function (a, b, c, d, e) {
        e = _super.prototype.as.call(this, a, b, c, d, e);
        var f = null;
        if (c.count > 0) {
            f = c._inner[0].r() + "[]";
        }
        else {
            var g = a.b.specificExternalType;
            f = CodeGenerationItemBuilder.o(this.k.platform, g, b, this.al, this.am) + "[]";
        }
        this.f.l("private " + f + " _" + CodeGeneratingComponentRenderer.e5(e) + " = null;");
        this.f.l("public " + f + " " + CodeGeneratingComponentRenderer.e6(e));
        this.f.l("{");
        this.f.h();
        this.f.l("get");
        this.f.l("{");
        this.f.h();
        this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(e) + " == null)");
        this.f.l("{");
        this.f.h();
        this.f.k("var " + CodeGeneratingComponentRenderer.e5(e) + " = new " + stringReplace(f, "[]", "") + "[" + c.count + "];");
        for (var h = 0; h < c.count; h++) {
            this.f.k(d._inner[h]);
            this.f.k(e + "[" + h + "] = " + c._inner[h].u + ";");
        }
        this.f.l("this._" + CodeGeneratingComponentRenderer.e5(e) + " = " + e + ";");
        this.f.f();
        this.f.l("}");
        this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(e) + ";");
        this.f.f();
        this.f.l("}");
        this.f.f();
        this.f.l("}");
        return CodeGeneratingComponentRenderer.e6(e);
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.i = function (a, b) {
        return new BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.get_y = function () {
        return false;
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.bv = function (a, b, c) {
        if (c.c == null) {
            return "null";
        }
        var d = c.c;
        var e = b;
        var f = new CodeGenerationItemBuilderPropertyInfo();
        f.b = ((function () {
            var $ret = new TypeDescriptionMetadata();
            $ret.propertyName = c.b.propertyName;
            $ret.specificType = b;
            $ret.specificExternalType = b;
            return $ret;
        })());
        f.b.knownType = 8;
        f.b.owningType = c.b.owningType;
        if (f.b.specificType == "string") {
            f.b.knownType = 2;
        }
        if (f.b.specificType == "int") {
            f.b.knownType = 1;
        }
        if (f.b.specificType == "double") {
            f.b.knownType = 1;
        }
        if (f.b.knownType == 8) {
            e = this.a7 + e;
        }
        var g = "@(new " + e + "[] { ";
        for (var h = 0; h < d.length; h++) {
            if (h > 0) {
                g += ", ";
            }
            f.c = d[h];
            if (f.b.knownType == 2) {
                g += "\"" + this.a3(a, f, true) + "\"";
            }
            else {
                g += this.a3(a, f, true);
            }
        }
        g += " })";
        return g;
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.at = function (a, b) {
        var c = b.b.specificType;
        if (c == null) {
            c = b.b.specificExternalType;
        }
        return this.bv(a, c, b);
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.bc = function (a, b, c) {
        _super.prototype.bc.call(this, a, b, c);
        this.h.l("<" + a + "");
        if (!this.bs.contains(this.ap)) {
            this.bs.add_1(this.ap);
            if (!this.bt.contains(this.aq)) {
                this.bt.add_1(this.aq);
                this.bb(this.aq);
            }
            this.f.l("private " + this.aq + " " + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
            if (!this.p) {
                if (this.v) {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                }
            }
            else {
                this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
            }
        }
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.bq = function () {
        this.h.l(">");
        this.h.h();
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.bl = function (a) {
        this.h.f();
        this.h.l("</" + a + ">");
        this.h.l("");
        _super.prototype.bl.call(this, a);
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.bi = function (a, b, c, d) {
        if (c.b != null && d.hasNameBinding(c.b.owningType, CodeGeneratingComponentRenderer.e6(c.b.propertyName)) && this.q(d, c)) {
            this.br(c);
            if (a > 0) {
                this.h.l("");
            }
            if (c.a == null || c.a.e == null) {
                this.h.j(b + "Name=\"@null\"");
            }
            else {
                this.h.j(b + "Name=\"" + c.a.e + "\"");
            }
            return;
        }
        if (this.s(c)) {
            this.br(c);
            if (!this.bs.contains(this.ap)) {
                this.bs.add_1(this.ap);
                if (!this.bt.contains(this.aq)) {
                    this.bt.add_1(this.aq);
                    this.bb(this.aq);
                }
                this.f.l("private " + this.aq + " " + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                if (!this.p) {
                    if (this.v) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                }
            }
            var e = CodeGeneratingComponentRenderer.e5(this.a3(d, c, true));
            var f = e;
            if (f != "null" && !stringStartsWith(f, "this.")) {
                f = "this." + f;
            }
            this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + "." + b + " = " + f + ";");
            return;
        }
        if (a > 0) {
            this.h.l("");
        }
        var g = false;
        if (b == "ComponentRenderer" && (c.b.owningType == "PropertyEditor" || c.b.owningType == "PropertyEditorPanel")) {
            g = true;
        }
        if (!g) {
            if (c.a != null && c.a.e != null) {
                if (this.m != null) {
                    if (this.m.library != null) {
                        if (this.m.library.hasItem(c.a.e)) {
                            var h = this.m.library.getItem(c.a.e);
                            var i = h.getContentForPlatform(5);
                            if (i != null && i.isJSContent) {
                                b = b + "Script";
                            }
                        }
                    }
                }
            }
            this.h.j(b);
            this.h.j("=");
            this.h.j("\"");
            this.h.j(this.a3(d, c, true));
            this.h.j("\"");
        }
        if (b == "Name") {
            this.h.l("");
            this.h.j("@ref=\"" + CodeGeneratingComponentRenderer.e5(this.a3(d, c, true)) + "\"");
        }
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.aw = function (a) {
        var b = a.b != null ? a.b.specificType + "." : "";
        return b + (a.c != null ? a.c.toString() : "");
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.a3 = function (a, b, c) {
        if (b.b == null) {
            return b.c;
        }
        if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
            b.c = this.ag(b);
        }
        switch (b.b.knownType) {
            case 11:
                {
                    var d = b.c;
                    return "@(new Point(" + d.x + ", " + d.y + "))";
                }
            case 14:
                {
                    var e = b.c;
                    return "@(new Rect(" + e.left + ", " + e.top + ", " + e.width + ", " + e.height + "))";
                }
            case 12:
                {
                    var f = b.c;
                    return "@(new Size(" + f.width + ", " + f.height + "))";
                }
            case 17:
                if (b.c == null) {
                    return "null";
                }
                if (typeof b.c === 'string') {
                    return "@(\"" + b.c.toString() + "\")";
                }
                if (typeCast(Boolean_$type, b.c) !== null) {
                    return b.c ? "true" : "false";
                }
                return b.c.toString();
        }
        return _super.prototype.a3.call(this, a, b, c);
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.ag = function (a) {
        var b = a.c;
        switch (b.a) {
            case 3: return b.b;
            case 0: return b.b;
            case 1: return b.b;
            case 2: return b.b;
        }
        return b.b;
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.a0 = function () {
        return "@null";
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.av = function (a) {
        return CodeGeneratingComponentRenderer.e5(CodeGeneratingComponentRenderer.e5(a.a.e));
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.bw = function (a) {
        a = stringReplace(a, "-", "_");
        return CodeGeneratingComponentRenderer.e6(a);
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.a1 = function (a) {
        this.br(a);
        return CodeGeneratingComponentRenderer.e6(a.a.e);
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.bj = function (a, b, c, d) {
        var e_2, _a;
        if (c.b.knownType == 9) {
            var e = typeCast(IList_$type, c.c);
            if (e != null) {
                try {
                    for (var _b = __values(fromEn(e)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var f = _c.value;
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            var g = f;
                            g.v(this, false);
                        }
                    }
                }
                catch (e_2_1) {
                    e_2 = { error: e_2_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_2)
                            throw e_2.error;
                    }
                }
            }
        }
        else {
            if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                var h = c.c;
                h.v(this, false);
            }
        }
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.prototype.get_a7 = function () {
        return "Igb";
    };
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.$t = markType(BlazorRazorCodeGeneratingComponentRendererCodeEmitter, 'BlazorRazorCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return BlazorRazorCodeGeneratingComponentRendererCodeEmitter;
}(CodeGeneratingComponentRendererCodeEmitter));
export { BlazorRazorCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter, _super);
    function BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b) {
        var _this = _super.call(this, a, b) || this;
        _this.bu = 0;
        _this.bs = new HashSet$1(String_$type, 0);
        _this.bt = new HashSet$1(String_$type, 0);
        return _this;
    }
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_r = function () {
        return true;
    };
    Object.defineProperty(BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype, "r", {
        get: function () {
            return this.get_r();
        },
        enumerable: false,
        configurable: true
    });
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.i = function (a, b) {
        return new BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_y = function () {
        return false;
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bv = function (a, b) {
        if (b.c == null) {
            return "null";
        }
        var c = b.c;
        var d = "new " + a + "[] { ";
        for (var e = 0; e < c.length; e++) {
            if (e > 0) {
                d += ", ";
            }
            d += "\"" + c[e] + "\"";
        }
        d += " }";
        return d;
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.at = function (a, b) {
        if (b.b.specificExternalType == "string") {
            return this.bv(b.b.specificExternalType, b);
        }
        if (b.c == null) {
            return "null";
        }
        var c = b.c;
        var d = "";
        for (var e = 0; e < c.length; e++) {
            if (e > 0) {
                d += ", ";
            }
            d += c[e];
        }
        return d;
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bc = function (a, b, c) {
        _super.prototype.bc.call(this, a, b, c);
        if (this.v) {
            this.h.l("var " + this.ap + " = new " + a + "();");
        }
        else {
            this.h.l("var " + this.ap + " = new " + a + "();");
        }
        if (!this.bs.contains(this.ap)) {
            this.bs.add_1(this.ap);
            if (!this.bt.contains(this.aq)) {
                this.bt.add_1(this.aq);
                this.bb(this.aq);
            }
            this.f.l("private " + this.aq + " " + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
            if (!this.p) {
                if (this.v) {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                }
            }
            else {
                this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
            }
        }
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bq = function () {
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bl = function (a) {
        _super.prototype.bl.call(this, a);
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bi = function (a, b, c, d) {
        if (c.b != null && d.hasNameBinding(c.b.owningType, CodeGeneratingComponentRenderer.e6(c.b.propertyName)) && this.q(d, c)) {
            this.br(c);
            if (a > 0) {
                this.h.l("");
            }
            if (c.a == null || c.a.e == null) {
                this.h.j(this.ap + "." + b + "Name = null;");
            }
            else {
                this.h.j(this.ap + "." + b + "Name = " + c.a.e + ";");
            }
            return;
        }
        if (this.s(c)) {
            this.br(c);
            if (!this.bs.contains(this.ap)) {
                this.bs.add_1(this.ap);
                if (!this.bt.contains(this.aq)) {
                    this.bt.add_1(this.aq);
                    this.bb(this.aq);
                }
                this.f.l("private " + this.aq + " " + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                if (!this.p) {
                    if (this.v) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                }
            }
            var e = CodeGeneratingComponentRenderer.e5(this.a3(d, c, true));
            var f = e;
            if (f != "null" && !stringStartsWith(f, "this.")) {
                f = "this." + f;
            }
            this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + "." + b + " = " + f + ";");
            return;
        }
        if (a > 0) {
            this.h.l("");
        }
        var g = false;
        if (b == "ComponentRenderer" && (c.b.owningType == "PropertyEditor" || c.b.owningType == "PropertyEditorPanel")) {
            g = true;
        }
        if (!g) {
            var h = false;
            if (c.a != null && c.a.e != null) {
                if (this.m != null) {
                    if (this.m.library != null) {
                        if (this.m.library.hasItem(c.a.e)) {
                            var i = this.m.library.getItem(c.a.e);
                            var j = i.getContentForPlatform(5);
                            if (j != null && j.isJSContent) {
                                b = b + "Script";
                                h = true;
                            }
                        }
                    }
                }
            }
            this.h.j(this.ap + ".");
            this.h.j(b);
            this.h.j(" = ");
            if (h) {
                this.h.j("\"" + this.a3(d, c, true) + "\"");
            }
            else {
                this.h.j(this.a3(d, c, true));
            }
            this.h.j(";");
        }
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.aw = function (a) {
        var b = a.b != null ? a.b.specificType + "." : "";
        return b + (a.c != null ? a.c.toString() : "");
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.a3 = function (a, b, c) {
        if (b.b == null) {
            return "\"" + b.c + "\"";
        }
        if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
            b.c = this.ag(b);
        }
        switch (b.b.knownType) {
            case 11:
                {
                    var d = b.c;
                    return "new Point(" + d.x + ", " + d.y + ")";
                }
            case 14:
                {
                    var e = b.c;
                    return "new Rect(" + e.left + ", " + e.top + ", " + e.width + ", " + e.height + ")";
                }
            case 12:
                {
                    var f = b.c;
                    return "new Size(" + f.width + ", " + f.height + ")";
                }
            case 17:
                if (b.c == null) {
                    return "null";
                }
                if (typeof b.c === 'string') {
                    return "\"" + b.c.toString() + "\"";
                }
                if (typeCast(Boolean_$type, b.c) !== null) {
                    return b.c ? "true" : "false";
                }
                return b.c.toString();
            case 5:
            case 6:
            case 16:
            case 2: return "\"" + _super.prototype.a3.call(this, a, b, c) + "\"";
            case 9: throw new NotSupportedException(1, "shouldn't get here");
        }
        return _super.prototype.a3.call(this, a, b, c);
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.ag = function (a) {
        var b = a.c;
        switch (b.a) {
            case 3: return b.b;
            case 0: return b.b;
            case 1: return b.b;
            case 2: return b.b;
        }
        return b.b;
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.a0 = function () {
        return "null";
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.av = function (a) {
        return CodeGeneratingComponentRenderer.e5(CodeGeneratingComponentRenderer.e5(a.a.e));
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bw = function (a) {
        a = stringReplace(a, "-", "_");
        return CodeGeneratingComponentRenderer.e6(a);
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.a1 = function (a) {
        this.br(a);
        return CodeGeneratingComponentRenderer.e6(a.a.e);
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bj = function (a, b, c, d) {
        var e_3, _a;
        if (c.b.knownType == 9) {
            var e = typeCast(IList_$type, c.c);
            if (e != null) {
                try {
                    for (var _b = __values(fromEn(e)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var f = _c.value;
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            var g = f;
                            g.v(this, false);
                            this.h.l(this.ap + "." + b + ".Add(" + this.a4 + ");");
                        }
                    }
                }
                catch (e_3_1) {
                    e_3 = { error: e_3_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_3)
                            throw e_3.error;
                    }
                }
            }
        }
        else if (c.b.knownType == 10) {
            var h = typeCast(Array_$type, c.c);
            if (h != null) {
                var i = "[";
                for (var j = 0; j < h.length; j++) {
                    var k = arrayGetValue(h, j);
                    if (typeCast(CodeGenerationItemBuilder.$, k) !== null) {
                        var l = k;
                        l.v(this, false);
                        i += this.a4 + ",";
                    }
                }
                i = trimEnd(i, ',');
                i += "]";
                this.h.l(this.ap + "." + b + " = " + i + ";");
            }
        }
        else {
            if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                var m = c.c;
                m.v(this, false);
                this.h.l(this.ap + "." + b + " = " + this.a4 + ";");
            }
        }
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_a7 = function () {
        return "Igb";
    };
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = markType(BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter;
}(CodeGeneratingComponentRendererCodeEmitter));
export { BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var ReactMarkupCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ReactMarkupCodeGeneratingComponentRendererCodeEmitter, _super);
    function ReactMarkupCodeGeneratingComponentRendererCodeEmitter(a, b) {
        var _this = _super.call(this, a, b) || this;
        _this.bs = new HashSet$1(String_$type, 0);
        _this.bt = new HashSet$1(String_$type, 0);
        return _this;
    }
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.get_r = function () {
        return false;
    };
    Object.defineProperty(ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype, "r", {
        get: function () {
            return this.get_r();
        },
        enumerable: false,
        configurable: true
    });
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.ar = function (a, b, c, d) {
        d = _super.prototype.ar.call(this, a, b, c, d);
        var e = b.r();
        this.f.l("private " + " _" + CodeGeneratingComponentRenderer.e5(d) + ": " + e + " | null = null;");
        this.f.l("public get " + CodeGeneratingComponentRenderer.e5(d) + "(): " + e + " {");
        this.f.h();
        this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(d) + " == null)");
        this.f.l("{");
        this.f.h();
        this.f.k(c);
        this.f.l("this._" + CodeGeneratingComponentRenderer.e5(d) + " = " + d + ";");
        this.f.f();
        this.f.l("}");
        this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(d) + ";");
        this.f.f();
        this.f.l("}");
        return CodeGeneratingComponentRenderer.e6(d);
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.as = function (a, b, c, d, e) {
        e = _super.prototype.as.call(this, a, b, c, d, e);
        var f = null;
        if (c.count > 0) {
            f = c._inner[0].r() + "[]";
        }
        else {
            var g = a.b.specificExternalType;
            f = CodeGenerationItemBuilder.o(this.k.platform, g, b, this.al, this.am) + "[]";
        }
        this.f.l("private" + " _" + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " | null = null;");
        this.f.l("public get " + CodeGeneratingComponentRenderer.e5(e) + "(): " + f + " {");
        this.f.h();
        this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(e) + " == null)");
        this.f.l("{");
        this.f.h();
        this.f.k("let " + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " = [];");
        for (var h = 0; h < c.count; h++) {
            this.f.k(d._inner[h]);
            this.f.k(e + ".push(" + c._inner[h].u + ")");
        }
        this.f.l("this._" + CodeGeneratingComponentRenderer.e5(e) + " = " + e + ";");
        this.f.f();
        this.f.l("}");
        this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(e) + ";");
        this.f.f();
        this.f.l("}");
        return CodeGeneratingComponentRenderer.e6(e);
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.i = function (a, b) {
        return new ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.at = function (a, b) {
        if (b.c == null) {
            return "{null}";
        }
        if (b.b != null && b.b.knownType == 23) {
            var c = b.c;
            var d = "";
            for (var e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d;
        }
        var f = b.c;
        var g = "[";
        for (var h = 0; h < f.length; h++) {
            if (h > 0) {
                g += ", ";
            }
            if (b.b != null && (b.b.specificExternalType == "int" || b.b.specificExternalType == "double")) {
                g += f[h].toString();
            }
            else {
                g += "\"" + f[h].toString() + "\"";
            }
        }
        g += "]";
        return g;
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.get_x = function () {
        return true;
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.get_a7 = function () {
        return "Igr";
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bc = function (a, b, c) {
        _super.prototype.bc.call(this, a, b, c);
        this.h.l("<" + this.bv(a) + "");
        this.h.h();
        if (!this.bt.contains(this.aq)) {
            this.bt.add_1(this.aq);
            this.bb(this.aq);
        }
        if (!this.bs.contains(this.ap)) {
            this.bs.add_1(this.ap);
            this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
            if (this.v) {
                this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + "Ref(r: " + this.aq + ") {");
                this.f.h();
                this.f.l("this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = r;");
                this.f.l("this.setState({});");
                this.f.f();
                this.f.l("}");
                this.g.l("this." + CodeGeneratingComponentRenderer.e5(this.ap) + "Ref = " + "this." + CodeGeneratingComponentRenderer.e5(this.ap) + "Ref.bind(this);");
            }
        }
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bv = function (a) {
        if (stringEndsWith(a, "Component")) {
            a = a.substr(0, a.length - ("Component").length);
        }
        return CodeGeneratingComponentRenderer.e6(a);
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bq = function () {
        if (!this.v && !this.p) {
            var a = this.bu();
            this.h.l("");
            this.h.l("name=\"" + a + "\"");
        }
        this.h.f();
        this.h.l(">");
        this.h.h();
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bu = function () {
        var a = CodeGeneratingComponentRenderer.e5(stringReplace(stringReplace(this.aq, this.a7, ""), "Component", ""));
        a = this.k.u(a);
        return a;
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.ax = function (a) {
        if (!stringIsNullOrEmpty(a.a.e)) {
            this.br(a);
            if (!this.bs.contains(a.a.e)) {
                this.g.l("this." + CodeGeneratingComponentRenderer.e5(a.a.e) + " = " + "this." + CodeGeneratingComponentRenderer.e5(a.a.e) + ".bind(this);");
            }
            return "this." + CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        return this.a0();
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bl = function (a) {
        this.h.f();
        this.h.l("</" + this.bv(a) + ">");
        _super.prototype.bl.call(this, a);
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bi = function (a, b, c, d) {
        var e = CodeGeneratingComponentRenderer.e5(b);
        var f = this.a3(d, c, true);
        if (c.b != null && d.hasNameBinding(c.b.owningType, CodeGeneratingComponentRenderer.e6(c.b.propertyName)) && this.q(d, c)) {
            this.br(c);
            if (a > 0) {
                this.h.l("");
            }
            if (c.a == null || c.a.e == null) {
                this.h.j(e + "Name={null}");
            }
            else {
                this.h.j(e + "Name=\"" + c.a.e + "\"");
            }
            return;
        }
        if (!(e == "name") || !this.v) {
            var g = (c.b != null && c.b.knownType == 10);
            if (this.q(d, c)) {
                this.br(c);
                var h = CodeGeneratingComponentRenderer.e5(f);
                var i = h;
                if (i != "null" && !stringStartsWith(i, "this.")) {
                    i = "this." + i;
                }
                f = "{" + i + "}";
            }
            else if (g) {
                f = "{" + f + "}";
            }
            if (a > 0) {
                this.h.l("");
            }
            this.h.j(e);
            this.h.j("=");
            if (!this.q(d, c) && !g) {
                this.h.j("\"");
            }
            this.h.j(f);
            if (!this.q(d, c) && !g) {
                this.h.j("\"");
            }
            a++;
        }
        if (e == "name") {
            if (this.v) {
                if (a > 0) {
                    this.h.l("");
                }
                this.h.j("ref={this." + CodeGeneratingComponentRenderer.e5(this.a3(d, c, true)) + "Ref}");
            }
        }
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.ag = function (a) {
        var b = a.c;
        switch (b.a) {
            case 3: return b.b;
            case 0: return b.b;
            case 1: return b.b;
            case 2: return b.b;
        }
        return b.b;
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.a0 = function () {
        return "null";
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.av = function (a) {
        return "this." + CodeGeneratingComponentRenderer.e5(a.a.e);
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.a1 = function (a) {
        return "this." + CodeGeneratingComponentRenderer.e5(a.a.e);
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bj = function (a, b, c, d) {
        var e_4, _a;
        if (c.b.knownType == 9) {
            var e = typeCast(IList_$type, c.c);
            if (e != null) {
                try {
                    for (var _b = __values(fromEn(e)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var f = _c.value;
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            var g = f;
                            g.v(this, false);
                        }
                    }
                }
                catch (e_4_1) {
                    e_4 = { error: e_4_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_4)
                            throw e_4.error;
                    }
                }
            }
        }
        else {
            if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                var h = c.c;
                h.v(this, false);
            }
        }
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bg = function (a, b, c) {
        _super.prototype.bg.call(this, a, b, c);
        WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
    };
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.$t = markType(ReactMarkupCodeGeneratingComponentRendererCodeEmitter, 'ReactMarkupCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return ReactMarkupCodeGeneratingComponentRendererCodeEmitter;
}(CodeGeneratingComponentRendererCodeEmitter));
export { ReactMarkupCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var WebCodeBehindCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebCodeBehindCodeGeneratingComponentRendererCodeEmitter, _super);
    function WebCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b) {
        var _this = _super.call(this, a, b) || this;
        _this.bv = 0;
        _this.bt = new HashSet$1(String_$type, 0);
        _this.bu = new HashSet$1(String_$type, 0);
        return _this;
    }
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_r = function () {
        return true;
    };
    Object.defineProperty(WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype, "r", {
        get: function () {
            return this.get_r();
        },
        enumerable: false,
        configurable: true
    });
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bs = function (a) {
        if (!a.b.owningContext.hasStringUnionPlatforms(a.b.owningType, a.b.propertyName)) {
            return false;
        }
        var b = a.b.owningContext.getStringUnionPlatforms(a.b.owningType, a.b.propertyName);
        for (var c = 0; c < b.length; c++) {
            if (b[c].toLowerCase() == this.a6) {
                return true;
            }
        }
        return false;
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.aw = function (a) {
        if (this.bs(a)) {
            return "\"" + CodeGeneratingComponentRenderer.e7(a.c.toString()) + "\"";
        }
        else {
            if (!this.bu.contains(a.b.specificType)) {
                this.bu.add_1(a.b.specificType);
                if (a.b.owningContext.getMetadata(a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)) != null) {
                    var b = a.b.owningContext.getMetadata(a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)).specificExternalType;
                    this.bb(a.b.specificType + ":" + b);
                }
                else if (a.b.owningContext.getMetadata("Web" + a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)) != null) {
                    var c = a.b.owningContext.getMetadata("Web" + a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)).specificExternalType;
                    this.bb(a.b.specificType + ":" + c);
                }
                else {
                    this.bb(a.b.specificType);
                }
            }
            var d = a.b != null ? a.b.specificType + "." : "";
            return d + (a.c != null ? a.c.toString() : "");
        }
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.at = function (a, b) {
        if (b.c == null) {
            return "null";
        }
        var c = b.c;
        var d = "[";
        for (var e = 0; e < c.length; e++) {
            if (e > 0) {
                d += ", ";
            }
            d += c[e];
        }
        return d + "]";
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_x = function () {
        return true;
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.u = function (a, b) {
        var c = false;
        var d = this.a6 == "react" ? b.getMetadata(a, "__isPlainObject") != null && (b.getMetadata(a, "__isTSPlainInterface") != null || b.getMetadata(a, "__marshalByValue") == null) : b.getMetadata(a, "__isPlainObject") != null;
        if (d) {
            c = true;
        }
        return c;
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bc = function (a, b, c) {
        _super.prototype.bc.call(this, a, b, c);
        var d = this.u(b, c);
        if (d) {
            if (this.v) {
                this.h.l("var " + this.ap + ": " + a + " = {} as " + a + ";");
            }
            else {
                this.h.l("var " + this.ap + ": " + a + " = {} as " + a + ";");
            }
        }
        else {
            if (this.v) {
                this.h.l("var " + this.ap + " = new " + a + "();");
            }
            else {
                this.h.l("var " + this.ap + " = new " + a + "();");
            }
        }
        if (!this.bt.contains(this.ap)) {
            this.bt.add_1(this.ap);
            if (!this.bu.contains(this.aq)) {
                this.bu.add_1(this.aq);
                this.bb(this.aq);
            }
            this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
            if (!this.p) {
                if (this.v) {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.querySelector('" + CodeGeneratingComponentRenderer.e8(stringReplace(this.aq, "Component", "")) + "') as " + this.aq + ";");
                }
            }
            else {
                this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.getElementById('" + this.ap + "') as " + this.aq + ";");
            }
        }
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bw = function (a) {
        if (stringEndsWith(a, "Component")) {
            a = a.substr(0, a.length - ("Component").length);
        }
        return CodeGeneratingComponentRenderer.e8(a);
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bq = function () {
        this.h.l("");
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bk = function (a) {
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bl = function (a) {
        {
            this.h.l("");
        }
        _super.prototype.bl.call(this, a);
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bi = function (a, b, c, d) {
        var e = CodeGeneratingComponentRenderer.e5(b);
        if (e == "name" && c.b == null) {
            return;
        }
        if (this.s(c)) {
            this.br(c);
            if (!this.bt.contains(this.ap)) {
                this.bt.add_1(this.ap);
                if (!this.bu.contains(this.aq)) {
                    this.bu.add_1(this.aq);
                    this.bb(this.aq);
                }
                this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
                if (!this.p) {
                    if (this.v) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.querySelector('" + CodeGeneratingComponentRenderer.e8(stringReplace(this.aq, "Component", "")) + "') as " + this.aq + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.getElementById('" + this.ap + "') as " + this.aq + ";");
                }
            }
            var f = CodeGeneratingComponentRenderer.e5(this.a3(d, c, true));
            var g = f;
            if (g != "null" && !stringStartsWith(g, "this.")) {
                g = "this." + g;
            }
            if (this.aq == "any" && stringContains(e, "-")) {
                this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + "['" + e + "'] = " + g + ";");
            }
            else {
                if (c.b != null && c.b.isCustomEvent) {
                    this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + ".addEventListener(\"" + e + "\", " + g + ");");
                }
                else {
                    this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + "." + e + " = " + g + ";");
                }
            }
            return;
        }
        if (a > 0) {
            this.h.l("");
        }
        if (this.aq == "any" && stringContains(e, "-")) {
            this.h.j(this.ap + "." + CodeGeneratingComponentRenderer.e5(this.ap) + "['" + e + "']");
        }
        else {
            this.h.j(this.ap + "." + CodeGeneratingComponentRenderer.e5(e));
        }
        this.h.j(" = ");
        this.h.j(this.a3(d, c, true));
        this.h.j(";");
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.ag = function (a) {
        var b = a.c;
        switch (b.a) {
            case 3: return b.b;
            case 0: return b.b;
            case 1: return b.b;
            case 2: return b.b;
        }
        return b.b;
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.a0 = function () {
        return "null";
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.av = function (a) {
        return CodeGeneratingComponentRenderer.e5(a.a.e);
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.a1 = function (a) {
        return CodeGeneratingComponentRenderer.e5(a.a.e);
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.ax = function (a) {
        if (!stringIsNullOrEmpty(a.a.e)) {
            this.br(a);
            if (!this.bt.contains(a.a.e)) {
                this.g.l("this." + CodeGeneratingComponentRenderer.e5(a.a.e) + " = " + "this." + CodeGeneratingComponentRenderer.e5(a.a.e) + ".bind(this);");
            }
            return "this." + CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        return this.a0();
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.a3 = function (a, b, c) {
        if (b.b == null) {
            return "\"" + b.c + "\"";
        }
        if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
            b.c = this.ag(b);
        }
        switch (b.b.knownType) {
            case 11:
                {
                    var d = b.c;
                    return "{ x: " + d.x + ", y: " + d.y + "}";
                }
            case 14:
                {
                    var e = b.c;
                    return "{ left: " + e.left + ", top: " + e.top + ", width: " + e.width + ", height: " + e.height + "}";
                }
            case 12:
                {
                    var f = b.c;
                    return "{ width: " + f.width + ", height: " + f.height + "}";
                }
            case 17:
                if (b.c == null) {
                    return "null";
                }
                if (typeof b.c === 'string') {
                    return "\"" + b.c.toString() + "\"";
                }
                if (typeCast(Boolean_$type, b.c) !== null) {
                    return b.c ? "true" : "false";
                }
                return b.c.toString();
            case 5:
            case 2: return "\"" + _super.prototype.a3.call(this, a, b, c) + "\"";
            case 23: return "[" + _super.prototype.a3.call(this, a, b, c) + "]";
            case 6: return "[\"" + stringJoin1(String_$type, "\", \"", _super.prototype.ab.call(this, b, c)) + "\"]";
            case 16: return "[\"" + stringJoin1(String_$type, "\", \"", _super.prototype.ac.call(this, b, c)) + "\"]";
            case 9: throw new NotSupportedException(1, "shouldn't get here");
        }
        return _super.prototype.a3.call(this, a, b, c);
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bj = function (a, b, c, d) {
        var e_5, _a;
        if (c.b.knownType == 9) {
            var e = typeCast(IList_$type, c.c);
            if (e != null) {
                var f = true;
                try {
                    for (var _b = __values(fromEn(e)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var g = _c.value;
                        if (typeCast(CodeGenerationItemBuilder.$, g) !== null) {
                            var h = g;
                            h.v(this, false);
                            var i = c.b.owningType;
                            var j = d.getMetadata(i, "__isPlainObject");
                            var k = false;
                            if (j != null) {
                                k = true;
                            }
                            if (k) {
                                if (f) {
                                    f = false;
                                    this.h.l(this.ap + "." + b + " = []");
                                }
                                this.h.l(this.ap + "." + b + ".push(" + this.a4 + ");");
                            }
                            else {
                                this.h.l(this.ap + "." + b + ".add(" + this.a4 + ");");
                            }
                        }
                    }
                }
                catch (e_5_1) {
                    e_5 = { error: e_5_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_5)
                            throw e_5.error;
                    }
                }
            }
        }
        else if (c.b.knownType == 10) {
            var l = typeCast(Array_$type, c.c);
            if (l != null) {
                var m = "[";
                for (var n = 0; n < l.length; n++) {
                    var o = arrayGetValue(l, n);
                    if (typeCast(CodeGenerationItemBuilder.$, o) !== null) {
                        var p = o;
                        p.v(this, false);
                        m += this.a4 + ",";
                    }
                }
                m = trimEnd(m, ',');
                m += "]";
                this.h.l(this.ap + "." + b + " = " + m + ";");
            }
        }
        else {
            if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                var q = c.c;
                q.v(this, false);
                this.h.l(this.ap + "." + b + " = " + this.a4 + ";");
            }
        }
    };
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = markType(WebCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'WebCodeBehindCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WebCodeBehindCodeGeneratingComponentRendererCodeEmitter;
}(CodeGeneratingComponentRendererCodeEmitter));
export { WebCodeBehindCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter, _super);
    function ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b) {
        return _super.call(this, a, b) || this;
    }
    ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_r = function () {
        return true;
    };
    ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.i = function (a, b) {
        return new ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
    };
    ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_a7 = function () {
        return "Igc";
    };
    ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bg = function (a, b, c) {
        _super.prototype.bg.call(this, a, b, c);
        WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
    };
    ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bx = function (a, b) {
        var c = false;
        if (b.getMetadata(a, "__isPlainObject") != null && b.getMetadata(a, "__marshalByValue") == null) {
            c = true;
        }
        return c;
    };
    ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = markType(ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter', WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$);
    return ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter;
}(WebCodeBehindCodeGeneratingComponentRendererCodeEmitter));
export { ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter, _super);
    function WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter(a, b) {
        var _this = _super.call(this, a, b) || this;
        _this.bv = 0;
        _this.bt = new HashSet$1(String_$type, 0);
        _this.bu = new HashSet$1(String_$type, 0);
        return _this;
    }
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.get_r = function () {
        return false;
    };
    Object.defineProperty(WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype, "r", {
        get: function () {
            return this.get_r();
        },
        enumerable: false,
        configurable: true
    });
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bs = function (a) {
        if (!a.b.owningContext.hasStringUnionPlatforms(a.b.owningType, a.b.propertyName)) {
            return false;
        }
        var b = a.b.owningContext.getStringUnionPlatforms(a.b.owningType, a.b.propertyName);
        for (var c = 0; c < b.length; c++) {
            if (b[c].toLowerCase() == this.a6) {
                return true;
            }
        }
        return false;
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.aw = function (a) {
        if (this.bs(a)) {
            return CodeGeneratingComponentRenderer.e7(a.c.toString());
        }
        else {
            return _super.prototype.aw.call(this, a);
        }
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.i = function (a, b) {
        return new WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.ar = function (a, b, c, d) {
        d = _super.prototype.ar.call(this, a, b, c, d);
        var e = b.r();
        this.f.l("private" + " _" + CodeGeneratingComponentRenderer.e5(d) + ": " + e + " | null = null;");
        this.f.l("public get " + CodeGeneratingComponentRenderer.e5(d) + "(): " + e + " {");
        this.f.h();
        this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(d) + " == null)");
        this.f.l("{");
        this.f.h();
        this.f.k(c);
        this.f.l("this._" + CodeGeneratingComponentRenderer.e5(d) + " = " + d + ";");
        this.f.f();
        this.f.l("}");
        this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(d) + ";");
        this.f.f();
        this.f.l("}");
        return CodeGeneratingComponentRenderer.e6(d);
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.as = function (a, b, c, d, e) {
        e = _super.prototype.as.call(this, a, b, c, d, e);
        var f = null;
        if (c.count > 0) {
            f = c._inner[0].r() + "[]";
        }
        else {
            var g = a.b.specificExternalType;
            f = CodeGenerationItemBuilder.o(this.k.platform, g, b, this.al, this.am) + "[]";
        }
        this.f.l("private" + " _" + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " | null = null;");
        this.f.l("public get " + CodeGeneratingComponentRenderer.e5(e) + "(): " + f + " {");
        this.f.h();
        this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(e) + " == null)");
        this.f.l("{");
        this.f.h();
        this.f.k("let " + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " = [];");
        for (var h = 0; h < c.count; h++) {
            this.f.k(d._inner[h]);
            this.f.k(e + ".push(" + c._inner[h].u + ")");
        }
        this.f.l("this._" + CodeGeneratingComponentRenderer.e5(e) + " = " + e + ";");
        this.f.f();
        this.f.l("}");
        this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(e) + ";");
        this.f.f();
        this.f.l("}");
        return CodeGeneratingComponentRenderer.e6(e);
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.at = function (a, b) {
        if (b.c == null) {
            return "null";
        }
        var c = b.c;
        var d = "";
        for (var e = 0; e < c.length; e++) {
            if (e > 0) {
                d += ", ";
            }
            d += c[e];
        }
        return d;
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.get_x = function () {
        return true;
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bc = function (a, b, c) {
        _super.prototype.bc.call(this, a, b, c);
        this.h.l("<" + this.bw(a) + "");
        if (!this.bt.contains(this.ap)) {
            this.bt.add_1(this.ap);
            if (!this.bu.contains(this.aq)) {
                this.bu.add_1(this.aq);
                this.bb(this.aq);
            }
            this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
            if (!this.p) {
                if (this.v) {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.querySelector('" + CodeGeneratingComponentRenderer.e8(stringReplace(this.aq, "Component", "")) + "') as " + this.aq + ";");
                }
            }
            else {
                this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.getElementById('" + this.ap + "') as " + this.aq + ";");
            }
        }
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bw = function (a) {
        if (stringEndsWith(a, "Component")) {
            a = a.substr(0, a.length - ("Component").length);
        }
        return CodeGeneratingComponentRenderer.e8(a);
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bq = function () {
        this.h.l(">");
        this.h.h();
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bl = function (a) {
        this.h.f();
        this.h.l("</" + this.bw(a) + ">");
        _super.prototype.bl.call(this, a);
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.get_a7 = function () {
        return "Igc";
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bg = function (a, b, c) {
        _super.prototype.bg.call(this, a, b, c);
        WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bi = function (a, b, c, d) {
        var e = CodeGeneratingComponentRenderer.e5(b);
        if (this.q(d, c)) {
            this.br(c);
            if (!this.bt.contains(this.ap)) {
                this.bt.add_1(this.ap);
                if (!this.bu.contains(this.aq)) {
                    this.bu.add_1(this.aq);
                    this.bb(this.aq);
                }
                this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
                if (!this.p) {
                    if (this.v) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.querySelector('" + CodeGeneratingComponentRenderer.e8(stringReplace(this.aq, "Component", "")) + "') as " + this.aq + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.getElementById('" + this.ap + "') as " + this.aq + ";");
                }
            }
            var f = CodeGeneratingComponentRenderer.e5(this.a3(d, c, true));
            var g = f;
            if (g != "null" && !stringStartsWith(g, "this.")) {
                g = "this." + g;
            }
            if (c.b.isCustomEvent) {
                this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + ".addEventListener(\"" + e + "\", " + g + ");");
            }
            else {
                this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + "." + e + " = " + g + ";");
            }
            return;
        }
        if (a > 0) {
            this.h.l("");
        }
        this.h.j(CodeGeneratingComponentRenderer.e8(e));
        this.h.j("=");
        this.h.j("\"");
        this.h.j(this.a3(d, c, true));
        this.h.j("\"");
        if (e == "name") {
            this.h.l("");
            this.h.j("id=\"" + this.a3(d, c, true) + "\"");
        }
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.ag = function (a) {
        var b = a.c;
        switch (b.a) {
            case 3: return b.b;
            case 0: return b.b;
            case 1: return b.b;
            case 2: return b.b;
        }
        return b.b;
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.a0 = function () {
        return "null";
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.av = function (a) {
        return CodeGeneratingComponentRenderer.e5(a.a.e);
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.a1 = function (a) {
        return CodeGeneratingComponentRenderer.e5(a.a.e);
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.ax = function (a) {
        if (!stringIsNullOrEmpty(a.a.e)) {
            this.br(a);
            if (!this.bt.contains(a.a.e)) {
                this.g.l("this." + CodeGeneratingComponentRenderer.e5(a.a.e) + " = " + "this." + CodeGeneratingComponentRenderer.e5(a.a.e) + ".bind(this);");
            }
            return "this." + CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        return this.a0();
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.prototype.bj = function (a, b, c, d) {
        var e_6, _a;
        if (c.b.knownType == 9) {
            var e = typeCast(IList_$type, c.c);
            if (e != null) {
                try {
                    for (var _b = __values(fromEn(e)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var f = _c.value;
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            var g = f;
                            g.v(this, false);
                        }
                    }
                }
                catch (e_6_1) {
                    e_6 = { error: e_6_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_6)
                            throw e_6.error;
                    }
                }
            }
        }
        else {
            if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                var h = c.c;
                h.v(this, false);
            }
        }
    };
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.$t = markType(WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter, 'WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter;
}(CodeGeneratingComponentRendererCodeEmitter));
export { WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter, _super);
    function WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b) {
        return _super.call(this, a, b) || this;
    }
    WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.i = function (a, b) {
        return new WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
    };
    WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_a7 = function () {
        return "Igc";
    };
    WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bg = function (a, b, c) {
        _super.prototype.bg.call(this, a, b, c);
        WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
    };
    WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = markType(WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter', WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$);
    return WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter;
}(WebCodeBehindCodeGeneratingComponentRendererCodeEmitter));
export { WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var TSCodeGeneratingComponentRendererHandlerEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TSCodeGeneratingComponentRendererHandlerEmitter, _super);
    function TSCodeGeneratingComponentRendererHandlerEmitter(a, b, c) {
        var _this = _super.call(this, a, b) || this;
        _this.w = 0;
        _this.x = new Dictionary$2(String_$type, String_$type, 0);
        _this.w = c;
        _this.x.item("IgxRowType", "RowType");
        _this.x.item("IgxCellType", "CellType");
        return _this;
    }
    TSCodeGeneratingComponentRendererHandlerEmitter.prototype.t = function (a, b) {
        _super.prototype.t.call(this, a, b);
        this.g.l("this." + CodeGeneratingComponentRenderer.e5(a) + "();");
    };
    TSCodeGeneratingComponentRendererHandlerEmitter.prototype.u = function (a, b) {
        _super.prototype.u.call(this, a, b);
        this.h.l("this." + CodeGeneratingComponentRenderer.e5(a) + "();");
    };
    Object.defineProperty(TSCodeGeneratingComponentRendererHandlerEmitter.prototype, "aa", {
        get: function () {
            switch (this.w) {
                case 1: return "angular";
                case 2: return "react";
                case 3: return "webcomponents";
            }
            return "";
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TSCodeGeneratingComponentRendererHandlerEmitter.prototype, "ab", {
        get: function () {
            switch (this.w) {
                case 1: return "Igx";
                case 2: return "Igr";
                case 3: return "Igc";
            }
            return "";
        },
        enumerable: false,
        configurable: true
    });
    TSCodeGeneratingComponentRendererHandlerEmitter.prototype.v = function (a, b) {
        var _this = this;
        _super.prototype.v.call(this, a, b);
        var c = new CodeGeneratingRegexHelper("import {([^}]+)}\\s+from\\s+['\"]([^'\"]+)['\"]");
        c.execute(b, function (d, e, f) {
            var g = d(1);
            var h = d(2);
            h = stringReplace(h, "webcomponents", _this.aa);
            h = stringReplace(h, "react", _this.aa);
            h = stringReplace(h, "angular", _this.aa);
            if (_this.aa == "angular" || _this.aa == "react") {
                if (stringContains(h, "/grids")) {
                    h = stringReplace(h, "/grids", "");
                }
            }
            if (_this.aa == "angular") {
                if (stringContains(h, "igniteui-angular-grids")) {
                    h = stringReplace(h, "igniteui-angular-grids", "igniteui-angular");
                }
            }
            var i = g.split(',');
            for (var k = 0; k < i.length; k++) {
                var j = i[k];
                var l = j.trim();
                if (stringStartsWith(l, "Igc")) {
                    l = _this.ab + l.substr(3);
                }
                if (stringStartsWith(l, "Igr")) {
                    l = _this.ab + l.substr(3);
                }
                if (stringStartsWith(l, "Igx")) {
                    l = _this.ab + l.substr(3);
                }
                if (stringStartsWith(l, "Igr") && stringEndsWith(l, "Component")) {
                    l = l.substr(0, l.length - ("Component").length);
                }
                if (h.trim() == "igniteui-angular") {
                    if (_this.x.containsKey(l.trim())) {
                        l = _this.x.item(l.trim());
                    }
                    if (stringStartsWith(l.trim(), "Igx") && stringEndsWith(l.trim(), "EventArgs")) {
                        _this.x.item(l.trim(), stringReplace(l, "Igx", "I"));
                        l = stringReplace(l, "Igx", "I");
                    }
                }
                a.addPackageImport(h.trim(), l.trim());
            }
            return null;
        });
    };
    TSCodeGeneratingComponentRendererHandlerEmitter.prototype.get_z = function () {
        switch (this.n.platform) {
            case 1: return "Component";
            case 2: return "";
            case 3: return "Component";
            case 4: return "";
            case 0: return "";
        }
        return "";
    };
    Object.defineProperty(TSCodeGeneratingComponentRendererHandlerEmitter.prototype, "z", {
        get: function () {
            return this.get_z();
        },
        enumerable: false,
        configurable: true
    });
    TSCodeGeneratingComponentRendererHandlerEmitter.prototype.get_y = function () {
        switch (this.n.platform) {
            case 1: return "Igx";
            case 2: return "Igr";
            case 3: return "Igc";
            case 4: return "Ultra";
            case 0: return "Xam";
            case 5: return "Igb";
        }
        return "";
    };
    Object.defineProperty(TSCodeGeneratingComponentRendererHandlerEmitter.prototype, "y", {
        get: function () {
            return this.get_y();
        },
        enumerable: false,
        configurable: true
    });
    TSCodeGeneratingComponentRendererHandlerEmitter.prototype.q = function (a, b) {
        return a;
    };
    TSCodeGeneratingComponentRendererHandlerEmitter.$t = markType(TSCodeGeneratingComponentRendererHandlerEmitter, 'TSCodeGeneratingComponentRendererHandlerEmitter', CodeGeneratingComponentRendererHandlerEmitter.$);
    return TSCodeGeneratingComponentRendererHandlerEmitter;
}(CodeGeneratingComponentRendererHandlerEmitter));
export { TSCodeGeneratingComponentRendererHandlerEmitter };
/**
 * @hidden
 */
var TSCodeGeneratingComponentRendererDataEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TSCodeGeneratingComponentRendererDataEmitter, _super);
    function TSCodeGeneratingComponentRendererDataEmitter(a, b) {
        var _this = _super.call(this, a, b) || this;
        _this.am = new Dictionary$2(CodeGeneratingCodeWriter.$, Number_$type, 0);
        return _this;
    }
    TSCodeGeneratingComponentRendererDataEmitter.prototype.x = function (a, b) {
        if (b === void 0) {
            b = false;
        }
        a = CodeGeneratingComponentRenderer.e6(a);
        a = stringReplace(a, "-", "_");
        if (b) {
            a += "Item";
        }
        return a;
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.w = function (a) {
        a = stringReplace(a, "-", "_");
        if (this.p.g) {
            return a;
        }
        return CodeGeneratingComponentRenderer.e5(a);
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.aq = function (a) {
        a = stringReplace(a, "-", "_");
        return "_" + CodeGeneratingComponentRenderer.e5(a);
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.get_t = function () {
        return ".ts";
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.al = function (a, b, c, d, e, f) {
        var e_7, _a;
        _super.prototype.al.call(this, a, b, c, d, e, f);
        var g = this.s(this.x(b, true), this.x(b, false), c, d);
        try {
            for (var _b = __values(fromEnum(g)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var h = _c.value;
                this.ad(h, e, f);
            }
        }
        catch (e_7_1) {
            e_7 = { error: e_7_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_7)
                    throw e_7.error;
            }
        }
        var i = 0;
        if (f == this.k) {
            f.addPackageImport("./" + this.x(b, false), this.x(b, false));
        }
        else {
            f.addPackageImport("./" + this.x(b, false), this.x(b, false));
        }
        e.l("export class " + this.x(b, false) + " extends Array<" + this.x(b, true) + "> {");
        e.h();
        e.l("public constructor(items: Array<" + this.x(b, true) + "> | number = -1) {");
        e.h();
        e.l("if (Array.isArray(items)) {");
        e.h();
        e.l("super(...items);");
        e.f();
        e.l("} else {");
        e.h();
        e.l("const newItems = [");
        e.h();
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.ah = function (a, b, c) {
        _super.prototype.ah.call(this, a, b, c);
        var d = this.p.cDNBasePath + this.x(a, false) + ".json";
        var e = "";
        if (this.n.platform == 2) {
            e = "this.setState({}); ";
        }
        if (this.n.platform == 3) {
            e = "if ((this as any)._bind) { (this as any)._bind(); } ";
        }
        if (this.n.platform == 1) {
            e = "this._detector.markForCheck(); ";
        }
        var f = this.x(a, false);
        var g = "null";
        switch (b) {
            case 3:
            case 2:
                f = "any[]";
                break;
            case 1:
                f = "any[]";
                g = this.x(a, false);
                break;
            case 0: break;
        }
        this.g.l("private " + this.aq(a) + ": " + f + " = " + g + ";");
        if ((b == 3 || b == 2) || (c != null && c.isAsync)) {
            this.g.l("private _isFetching: boolean = false;");
        }
        this.g.l("public get " + CodeGeneratingComponentRenderer.e5(a) + "(): " + f + " {");
        this.g.h();
        if (c != null && c.isAsync) {
            this.g.l("if (this." + this.aq(a) + " == null && !this._isFetching)");
            this.g.l("{");
            this.g.h();
            this.g.l("this._isFetching = true;");
            this.g.l("( async () => { this." + this.aq(a) + " = await (await " + this.x(a, false) + ".fetch()); " + e + " })();");
            this.g.f();
            this.g.l("}");
        }
        else {
            if ((b == 3 || b == 2)) {
                this.g.l("if (this." + this.aq(a) + " == null && !this._isFetching)");
                this.g.l("{");
                this.g.h();
                this.g.l("this._isFetching = true;");
                this.g.l("( async () => {");
                this.g.l("    const url = '" + d + "';");
                this.g.l("    this." + this.aq(a) + " = await (await fetch(url)).json();");
                this.g.l("    " + e);
                this.g.l("})();");
                this.g.f();
                this.g.l("}");
            }
            if (b == 0) {
                this.g.l("if (this." + this.aq(a) + " == null)");
                this.g.l("{");
                this.g.h();
                this.g.l("this." + this.aq(a) + " = new " + this.x(a, false) + "();");
                this.g.f();
                this.g.l("}");
            }
        }
        this.g.l("return this." + this.aq(a) + ";");
        this.g.f();
        this.g.l("}");
        this.g.l("");
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.ai = function (a, b) {
        _super.prototype.ai.call(this, a, b);
        this.g.l("private " + this.aq(a) + ": " + this.an(b.e) + " = " + this.ao(b) + ";");
        this.g.l("public get " + CodeGeneratingComponentRenderer.e5(a) + "(): " + this.an(b.e) + " {");
        this.g.h();
        this.g.l("return this." + this.aq(a) + ";");
        this.g.f();
        this.g.l("}");
        this.g.l("");
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.aj = function (a, b, c, d) {
        _super.prototype.aj.call(this, a, b, c, d);
        this.g.l("private " + this.aq(a) + ": " + this.x(a, false) + "Item = null;");
        this.g.l("public get " + CodeGeneratingComponentRenderer.e5(a) + "(): " + this.x(a, false) + "Item {");
        this.g.h();
        this.g.l("if (this." + this.aq(a) + " == null)");
        this.g.l("{");
        this.g.h();
        this.g.l("this." + this.aq(a) + " = ");
        this.ar(this.x(a, true), b, c, d, this.g);
        this.g.l("}");
        this.g.l("return this." + this.aq(a) + ";");
        this.g.f();
        this.g.l("}");
        this.g.l("");
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.z = function (a, b) {
        b.addPackageImport("./" + a, a);
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.aa = function (a, b, c) {
        c.addDefaultImport("./" + b, a);
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.ab = function (a, b) {
        this.l = CodeGeneratingImportManager.merge(this.l, b);
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.ad = function (a, b, c) {
        var d = 0;
        if (c == this.k) {
            c.addPackageImport("./" + a.c, a.d);
        }
        else {
            c.addPackageImport("./" + a.c, a.d);
        }
        b.l("export class " + a.d + " {");
        b.h();
        b.l("public constructor(init: Partial<" + a.d + ">) {");
        b.h();
        b.l("Object.assign(this, init);");
        b.f();
        b.l("}");
        b.l("");
        for (var e = 0; e < a.a.length; e++) {
            var f = a.a[e];
            var g = a.b[e];
            b.l("public " + this.w(f) + ": " + this.ap(a, f, g) + ";");
        }
        b.f();
        b.l("");
        b.l("}");
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.an = function (a) {
        switch (a) {
            case 1: return "boolean";
            case 0: return "number";
            case 2: return "string";
            case 3: return "any";
        }
        return "any";
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.ap = function (a, b, c) {
        if (c.d) {
            return a.d + "_" + CodeGeneratingComponentRenderer.e6(this.w(b)) + "Item[]";
        }
        else if (c.f) {
            return a.d + "_" + TypeDescriptionContext.ab(this.w(b));
        }
        else {
            switch (c.c) {
                case 1: return "boolean";
                case 0: return "number";
                case 2: return "string";
                case 3: return "any";
            }
        }
        return "any";
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.ak = function (a, b, c, d, e) {
        _super.prototype.ak.call(this, a, b, c, d, e);
        e.f();
        e.l("];");
        e.l("super(...newItems.slice(0));");
        e.f();
        e.l("}");
        e.f();
        e.l("}");
        e.f();
        e.l("}");
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.y = function (a, b, c, d, e, f, g) {
        _super.prototype.y.call(this, a, b, c, d, e, f, g);
        this.ar(this.x(b, true), c, d, e, g);
        g.l(",");
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.ar = function (a, b, c, d, e) {
        if (d == null) {
            e.j("null");
            return;
        }
        if (typeCast(JsonDictionaryArray.$, d) !== null) {
            var f = d;
            e.l("[");
            e.h();
            if (f.items != null && f.items.length > 0) {
                for (var g = 0; g < f.items.length; g++) {
                    if (g > 0) {
                        e.l(",");
                    }
                    this.ar(a, b, c, f.items[g], e);
                }
            }
            e.f();
            e.l("]");
        }
        else if (typeCast(JsonDictionaryObject.$, d) !== null) {
            var h = d;
            e.l("new " + a + "(");
            e.l("{");
            e.h();
            var i = true;
            for (var j = 0; j < b.length; j++) {
                var k = b[j];
                var l = c[j];
                if (!h.g(k)) {
                    continue;
                }
                if (i) {
                    i = false;
                }
                else {
                    e.l(",");
                }
                e.j(this.w(k) + ": ");
                if (l.d) {
                    var m = a + "_" + CodeGeneratingComponentRenderer.e6(this.w(k)) + "Item";
                    if (h.g(k)) {
                        this.ar(m, l.a, l.b, h.item(k), e);
                    }
                    else {
                        e.j("[]");
                    }
                }
                else if (l.f) {
                    var n = a + "_" + CodeGeneratingComponentRenderer.e6(this.w(k));
                    this.ar(n, l.a, l.b, h.item(k), e);
                }
                else {
                    var o = h.item(k);
                    e.j(this.u(l, o));
                }
            }
            e.l("");
            e.f();
            e.j("})");
        }
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.u = function (a, b) {
        if (b.value == null) {
            return ("null");
        }
        switch (a.c) {
            case 1: return (b.value ? "true" : "false");
            case 0: return (b.value.toString());
            case 2: return ("`" + stringReplace(b.value, "`", "\\`") + "`");
            case 3: return ("null");
        }
        return "";
    };
    TSCodeGeneratingComponentRendererDataEmitter.prototype.ao = function (a) {
        switch (a.e) {
            case 1: return (a.value ? "true" : "false");
            case 0: return (a.value.toString());
            case 2: return ("`" + a.value + "`");
            case 3: return "null";
        }
        return "";
    };
    TSCodeGeneratingComponentRendererDataEmitter.$t = markType(TSCodeGeneratingComponentRendererDataEmitter, 'TSCodeGeneratingComponentRendererDataEmitter', CodeGeneratingComponentRendererDataEmitter.$);
    return TSCodeGeneratingComponentRendererDataEmitter;
}(CodeGeneratingComponentRendererDataEmitter));
export { TSCodeGeneratingComponentRendererDataEmitter };
/**
 * @hidden
 */
var AngularTemplateCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AngularTemplateCodeGeneratingComponentRendererCodeEmitter, _super);
    function AngularTemplateCodeGeneratingComponentRendererCodeEmitter(a, b) {
        var _this = _super.call(this, a, b) || this;
        _this.bu = 0;
        _this.bs = new HashSet$1(String_$type, 0);
        _this.bt = new HashSet$1(String_$type, 0);
        return _this;
    }
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.get_r = function () {
        return false;
    };
    Object.defineProperty(AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype, "r", {
        get: function () {
            return this.get_r();
        },
        enumerable: false,
        configurable: true
    });
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.ar = function (a, b, c, d) {
        d = _super.prototype.ar.call(this, a, b, c, d);
        var e = b.r();
        this.f.l("private " + " _" + CodeGeneratingComponentRenderer.e5(d) + ": " + e + " | null = null;");
        this.f.l("public get " + CodeGeneratingComponentRenderer.e5(d) + "(): " + e + " {");
        this.f.h();
        this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(d) + " == null)");
        this.f.l("{");
        this.f.h();
        this.f.k(c);
        this.f.l("this._" + CodeGeneratingComponentRenderer.e5(d) + " = " + d + ";");
        this.f.f();
        this.f.l("}");
        this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(d) + ";");
        this.f.f();
        this.f.l("}");
        return CodeGeneratingComponentRenderer.e5(d);
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.as = function (a, b, c, d, e) {
        e = _super.prototype.as.call(this, a, b, c, d, e);
        var f = null;
        if (c.count > 0) {
            f = c._inner[0].r() + "[]";
        }
        else {
            var g = a.b.specificExternalType;
            f = CodeGenerationItemBuilder.o(this.k.platform, g, b, this.al, this.am) + "[]";
        }
        this.f.l("private" + " _" + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " | null = null;");
        this.f.l("public get " + CodeGeneratingComponentRenderer.e5(e) + "(): " + f + " {");
        this.f.h();
        this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(e) + " == null)");
        this.f.l("{");
        this.f.h();
        this.f.k("let " + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " = [];");
        for (var h = 0; h < c.count; h++) {
            this.f.k(d._inner[h]);
            this.f.k(e + ".push(" + c._inner[h].u + ")");
        }
        this.f.l("this._" + CodeGeneratingComponentRenderer.e5(e) + " = " + e + ";");
        this.f.f();
        this.f.l("}");
        this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(e) + ";");
        this.f.f();
        this.f.l("}");
        return CodeGeneratingComponentRenderer.e6(e);
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.i = function (a, b) {
        return new AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.at = function (a, b) {
        if (b.c == null) {
            return "null";
        }
        var c = b.c;
        var d = "";
        for (var e = 0; e < c.length; e++) {
            if (e > 0) {
                d += ", ";
            }
            d += c[e];
        }
        return d;
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.get_x = function () {
        return true;
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.get_a7 = function () {
        return "Igx";
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.bc = function (a, b, c) {
        _super.prototype.bc.call(this, a, b, c);
        this.h.l("<" + this.bv(a) + "");
        if (!this.bs.contains(this.ap)) {
            this.bs.add_1(this.ap);
            if (!this.bt.contains(this.aq)) {
                this.bt.add_1(this.aq);
                this.bb(this.aq);
            }
            this.f.l("@ViewChild(\"" + CodeGeneratingComponentRenderer.e5(this.ap) + "\", { static: true } )");
            this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
        }
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.bg = function (a, b, c) {
        _super.prototype.bg.call(this, a, b, c);
        WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.bv = function (a) {
        if (stringEndsWith(a, "Component")) {
            a = a.substr(0, a.length - ("Component").length);
        }
        return CodeGeneratingComponentRenderer.e8(a);
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.bq = function () {
        this.h.l(">");
        this.h.h();
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.bl = function (a) {
        this.h.f();
        this.h.l("</" + this.bv(a) + ">");
        _super.prototype.bl.call(this, a);
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.bi = function (a, b, c, d) {
        var e = CodeGeneratingComponentRenderer.e5(b);
        if (this.q(d, c)) {
            this.br(c);
            if (this.t(c)) {
                e = "(" + e + ")";
            }
            else {
                e = "[" + e + "]";
            }
        }
        if (a > 0) {
            this.h.l("");
        }
        this.h.j(e);
        this.h.j("=");
        this.h.j("\"");
        this.h.j(this.a3(d, c, true));
        this.h.j("\"");
        if (e == "name") {
            this.h.l("");
            this.h.j("#" + CodeGeneratingComponentRenderer.e5(this.a3(d, c, true)));
        }
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.ax = function (a) {
        if (!stringIsNullOrEmpty(a.a.e)) {
            this.br(a);
            return "this." + CodeGeneratingComponentRenderer.e5(a.a.e) + "($event)";
        }
        return this.a0();
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.ag = function (a) {
        var b = a.c;
        switch (b.a) {
            case 3: return b.b;
            case 0: return b.b;
            case 1: return b.b;
            case 2: return b.b;
        }
        return b.b;
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.a0 = function () {
        return "null";
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.av = function (a) {
        return CodeGeneratingComponentRenderer.e5(a.a.e);
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.a1 = function (a) {
        return CodeGeneratingComponentRenderer.e5(a.a.e);
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.prototype.bj = function (a, b, c, d) {
        var e_8, _a;
        if (c.b.knownType == 9) {
            var e = typeCast(IList_$type, c.c);
            if (e != null) {
                try {
                    for (var _b = __values(fromEn(e)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var f = _c.value;
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            var g = f;
                            g.v(this, false);
                        }
                    }
                }
                catch (e_8_1) {
                    e_8 = { error: e_8_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_8)
                            throw e_8.error;
                    }
                }
            }
        }
        else {
            if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                var h = c.c;
                h.v(this, false);
            }
        }
    };
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.$t = markType(AngularTemplateCodeGeneratingComponentRendererCodeEmitter, 'AngularTemplateCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return AngularTemplateCodeGeneratingComponentRendererCodeEmitter;
}(CodeGeneratingComponentRendererCodeEmitter));
export { AngularTemplateCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter, _super);
    function AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b) {
        return _super.call(this, a, b) || this;
    }
    AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_r = function () {
        return true;
    };
    AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.i = function (a, b) {
        return new AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
    };
    AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_a7 = function () {
        return "Igc";
    };
    AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.bg = function (a, b, c) {
        _super.prototype.bg.call(this, a, b, c);
        WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
    };
    AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = markType(AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter', WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$);
    return AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter;
}(WebCodeBehindCodeGeneratingComponentRendererCodeEmitter));
export { AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var DotNetCodeGeneratingComponentRendererDataEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DotNetCodeGeneratingComponentRendererDataEmitter, _super);
    function DotNetCodeGeneratingComponentRendererDataEmitter(a, b) {
        var _this = _super.call(this, a, b) || this;
        _this.am = false;
        return _this;
    }
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.ab = function (a, b) {
        _super.prototype.ab.call(this, a, b);
        this.f.l("<Compile Include=\"" + a + ".cs\">");
        this.f.h();
        this.f.l("<SubType>Code</SubType>");
        this.f.f();
        this.f.l("</Compile>");
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.af = function (a) {
        a.l("using System;");
        a.l("using System.Collections.Generic;");
        if (!this.am) {
            this.am = true;
        }
        _super.prototype.af.call(this, a);
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.ae = function (a) {
        _super.prototype.ae.call(this, a);
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.x = function (a, b) {
        if (b === void 0) {
            b = false;
        }
        a = CodeGeneratingComponentRenderer.e6(a);
        a = stringReplace(a, "-", "_");
        if (b) {
            a += "Item";
        }
        return a;
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.w = function (a) {
        a = stringReplace(a, "-", "_");
        return CodeGeneratingComponentRenderer.e6(a);
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.aq = function (a) {
        a = stringReplace(a, "-", "_");
        return "_" + CodeGeneratingComponentRenderer.e5(a);
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.al = function (a, b, c, d, e, f) {
        var e_9, _a;
        _super.prototype.al.call(this, a, b, c, d, e, f);
        var g = this.s(this.x(b, true), this.x(b, false), c, d);
        try {
            for (var _b = __values(fromEnum(g)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var h = _c.value;
                this.ad(h, e, f);
            }
        }
        catch (e_9_1) {
            e_9 = { error: e_9_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_9)
                    throw e_9.error;
            }
        }
        e.l("");
        e.l("public class " + this.x(b, false));
        e.h();
        e.l(": List<" + this.x(b, true) + ">");
        e.f();
        e.l("{");
        e.h();
        e.l("public " + this.x(b, false) + "()");
        e.l("{");
        e.h();
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.ah = function (a, b, c) {
        _super.prototype.ah.call(this, a, b, c);
        if (b == 3) {
            var d = this.p.cDNBasePath + this.x(a, false) + ".json";
            this.g.l("public RemoteJSON " + this.w(a));
            this.g.l("{");
            this.g.h();
            this.g.l("get");
            this.g.l("{");
            this.g.h();
            this.g.l("return RemoteJSON.LoadFrom(\"" + d + "\");");
            this.g.f();
            this.g.l("}");
            this.g.l("");
            return;
        }
        var e = this.aq(a) + " = " + "new " + this.x(a, false) + "(); ";
        var f = "";
        if (this.n.platform == 5) {
            f = " StateHasChanged();";
        }
        else {
            f = " OnPropertyChanged(\"" + this.x(a, false) + "\");";
        }
        if (c != null && (b == 2 || c.isAsync)) {
            e = this.x(a, false) + ".Fetch().ContinueWith((t) => {" + this.aq(a) + " = t.Result; " + f + " });";
        }
        this.g.l("private " + this.x(a, false) + " " + this.aq(a) + " = null;");
        this.g.l("public " + this.x(a, false) + " " + this.w(a));
        this.g.l("{");
        this.g.h();
        this.g.l("get");
        this.g.l("{");
        this.g.h();
        this.g.l("if (" + this.aq(a) + " == null)");
        this.g.l("{");
        this.g.h();
        this.g.l(e);
        this.g.f();
        this.g.l("}");
        this.g.l("return " + this.aq(a) + ";");
        this.g.f();
        this.g.l("}");
        this.g.f();
        this.g.l("}");
        this.g.l("");
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.ai = function (a, b) {
        _super.prototype.ai.call(this, a, b);
        this.g.l("private " + this.an(b.e) + " " + this.aq(a) + " = " + this.ao(b) + ";");
        this.g.l("public " + this.an(b.e) + " " + this.w(a));
        this.g.l("{");
        this.g.h();
        this.g.l("get");
        this.g.l("{");
        this.g.h();
        this.g.l("return " + this.aq(a) + ";");
        this.g.f();
        this.g.l("}");
        this.g.f();
        this.g.l("}");
        this.g.l("");
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.aj = function (a, b, c, d) {
        _super.prototype.aj.call(this, a, b, c, d);
        this.g.l("private " + this.x(a, false) + "Item " + this.aq(a) + " = null;");
        this.g.l("public " + this.x(a, false) + "Item " + this.w(a));
        this.g.l("{");
        this.g.h();
        this.g.l("get");
        this.g.l("{");
        this.g.h();
        this.g.l("if (" + this.aq(a) + " == null)");
        this.g.l("{");
        this.g.h();
        this.g.l("_" + a + " = ");
        this.ar(this.x(a, true), b, c, d, this.g);
        this.g.f();
        this.g.l("}");
        this.g.l("return " + this.aq(a) + ";");
        this.g.f();
        this.g.l("}");
        this.g.f();
        this.g.l("}");
        this.g.l("");
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.get_t = function () {
        return ".cs";
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.ad = function (a, b, c) {
        b.l("public class " + a.d);
        b.l("{");
        b.h();
        for (var d = 0; d < a.a.length; d++) {
            var e = a.a[d];
            var f = a.b[d];
            b.l("public " + this.ap(a, e, f) + " " + this.w(e) + " { get; set; }");
        }
        b.f();
        b.l("}");
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.an = function (a) {
        switch (a) {
            case 1: return "bool";
            case 0: return "double";
            case 2: return "string";
            case 3: return "object";
        }
        return "object";
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.get_r = function () {
        return true;
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.ap = function (a, b, c) {
        if (c.d) {
            return "List<" + a.d + "_" + this.w(b) + "Item>";
        }
        else if (c.f) {
            return a.d + "_" + this.w(b);
        }
        else {
            switch (c.c) {
                case 1: return "bool";
                case 0: return "double";
                case 2: return "string";
                case 3: return "object";
            }
        }
        return "object";
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.ak = function (a, b, c, d, e) {
        _super.prototype.ak.call(this, a, b, c, d, e);
        e.f();
        e.l("}");
        if (a == 2 || a == 3) {
            var f = this.p.cDNBasePath + this.x(b, false) + ".json";
            e.l("public async static System.Threading.Tasks.Task<" + this.x(b, false) + "> Fetch()");
            e.l("{");
            e.h();
            e.l("var url = \"" + f + "\" ;");
            e.l("var client = new System.Net.Http.HttpClient();");
            e.l("var str = await client.GetStringAsync(url);");
            e.l("var arr = System.Text.Json.JsonSerializer.Deserialize<" + this.x(b, false) + ">(str);");
            e.l("return arr;");
            e.f();
            e.l("}");
            e.l("");
        }
        e.f();
        e.l("}");
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.y = function (a, b, c, d, e, f, g) {
        _super.prototype.y.call(this, a, b, c, d, e, f, g);
        if (a == 2 || a == 3) {
            return;
        }
        g.j("this.Add(");
        this.ar(this.x(b, true), c, d, e, g);
        g.l(");");
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.ar = function (a, b, c, d, e) {
        if (d == null) {
            e.j("null");
            return;
        }
        if (typeCast(JsonDictionaryArray.$, d) !== null) {
            var f = d;
            e.l("new List<" + a + ">()");
            e.l("{");
            e.h();
            if (f.items != null && f.items.length > 0) {
                for (var g = 0; g < f.items.length; g++) {
                    if (g > 0) {
                        e.l(",");
                    }
                    this.ar(a, b, c, f.items[g], e);
                }
            }
            e.f();
            e.l("}");
        }
        else if (typeCast(JsonDictionaryObject.$, d) !== null) {
            var h = d;
            e.l("new " + a + "()");
            e.l("{");
            e.h();
            for (var i = 0; i < b.length; i++) {
                var j = b[i];
                var k = c[i];
                if (!h.g(j)) {
                    continue;
                }
                if (i > 0) {
                    e.l(",");
                }
                e.j(j + " = ");
                if (k.d) {
                    var l = a + "_" + this.w(j) + "Item";
                    this.ar(l, k.a, k.b, h.item(j), e);
                }
                else if (k.f) {
                    var m = a + "_" + this.w(j);
                    this.ar(m, k.a, k.b, h.item(j), e);
                }
                else {
                    var n = h.item(j);
                    try {
                        e.j(this.u(k, n));
                    }
                    catch (o) {
                        var p = getInstanceType(n.value).typeName;
                        var q = stringReplace(EnumUtil.getName(JsonDictionaryValueType_$type, k.c), "Value", "");
                        var r = "'" + j + "' property (" + n.value + ") is '" + p + "' instead of '" + q + "' type.";
                        throw new BaseError(1, "Failed on creating '" + a + "' because " + r + "\n" + o.message);
                    }
                }
            }
            e.l("");
            e.f();
            e.j("}");
        }
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.u = function (a, b) {
        switch (a.c) {
            case 1: return (b.value ? "true" : "false");
            case 0:
                if (b == null || b.value == null) {
                    return "double.NaN";
                }
                return (b.value.toString());
            case 2: if (b.value == null) {
                return "null";
            }
            else {
                return ("@\"" + stringReplace(b.value, "\"", "\"\"") + "\"");
            }
            case 3: return ("null");
        }
        return "";
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.prototype.ao = function (a) {
        switch (a.e) {
            case 1: return (a.value ? "true" : "false");
            case 0: return (a.value.toString());
            case 2: return ("@\"" + stringReplace(a.value, "\"", "\"\"") + "\"");
            case 3: return "null";
        }
        return "";
    };
    DotNetCodeGeneratingComponentRendererDataEmitter.$t = markType(DotNetCodeGeneratingComponentRendererDataEmitter, 'DotNetCodeGeneratingComponentRendererDataEmitter', CodeGeneratingComponentRendererDataEmitter.$);
    return DotNetCodeGeneratingComponentRendererDataEmitter;
}(CodeGeneratingComponentRendererDataEmitter));
export { DotNetCodeGeneratingComponentRendererDataEmitter };
/**
 * @hidden
 */
var WPFXamlCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WPFXamlCodeGeneratingComponentRendererCodeEmitter, _super);
    function WPFXamlCodeGeneratingComponentRendererCodeEmitter(a, b) {
        return _super.call(this, a, b) || this;
    }
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.a9 = function (a) {
        if (a == null || isNaN_(a)) {
            return "";
        }
        return timeSpanFromMilliseconds(a).toString();
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.get_r = function () {
        return false;
    };
    Object.defineProperty(WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype, "r", {
        get: function () {
            return this.get_r();
        },
        enumerable: false,
        configurable: true
    });
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.i = function (a, b) {
        return new WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.get_y = function () {
        return false;
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.at = function (a, b) {
        if (b.c == null) {
            return "null";
        }
        var c = b.c;
        var d = "";
        for (var e = 0; e < c.length; e++) {
            if (e > 0) {
                d += ", ";
            }
            d += c[e];
        }
        return d;
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.bc = function (a, b, c) {
        _super.prototype.bc.call(this, a, b, c);
        this.h.l("<ig:" + a + "");
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.bq = function () {
        this.h.l(">");
        this.h.h();
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.bl = function (a) {
        this.h.f();
        this.h.l("</ig:" + a + ">");
        _super.prototype.bl.call(this, a);
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.bi = function (a, b, c, d) {
        if (b == "Label" && c.b != null && stringContains(c.b.owningType, "Axis")) {
            if (c.c != null && stringContains(c.c, "{")) {
                if (a > 0) {
                    this.h.l("");
                }
                this.h.j(b);
                this.h.j("=");
                this.h.j("\"");
                this.h.j("{}" + this.a3(d, c, true));
                this.h.j("\"");
                return;
            }
        }
        if (a > 0) {
            this.h.l("");
        }
        this.h.j(b);
        this.h.j("=");
        this.h.j("\"");
        this.h.j(this.a3(d, c, true));
        this.h.j("\"");
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.ag = function (a) {
        var b = a.c;
        switch (b.a) {
            case 3: return b.b;
            case 0: return b.b;
            case 1: return b.b;
            case 2: return b.b;
        }
        return b.b;
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.a0 = function () {
        return "{x:Null}";
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.av = function (a) {
        return "{Binding ElementName=" + a.a.e + "}";
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.bs = function (a) {
        a = stringReplace(a, "-", "_");
        return CodeGeneratingComponentRenderer.e6(a);
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.a1 = function (a) {
        this.br(a);
        return "{Binding " + this.bs(a.a.e) + "}";
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.bj = function (a, b, c, d) {
        var e_10, _a;
        this.h.j("<ig:");
        {
            this.h.j(this.aq);
            this.h.j(".");
        }
        this.h.j(b);
        this.h.l(">");
        this.h.h();
        if (c.b.knownType == 9) {
            var e = typeCast(IList_$type, c.c);
            if (e != null) {
                try {
                    for (var _b = __values(fromEn(e)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var f = _c.value;
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            var g = f;
                            g.v(this, false);
                        }
                    }
                }
                catch (e_10_1) {
                    e_10 = { error: e_10_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_10)
                            throw e_10.error;
                    }
                }
            }
        }
        else {
            if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                var h = c.c;
                h.v(this, false);
            }
        }
        this.h.f();
        this.h.j("</ig:");
        {
            this.h.j(this.aq);
            this.h.j(".");
        }
        this.h.j(b);
        this.h.l(">");
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.prototype.get_a7 = function () {
        return "Xam";
    };
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.$t = markType(WPFXamlCodeGeneratingComponentRendererCodeEmitter, 'WPFXamlCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WPFXamlCodeGeneratingComponentRendererCodeEmitter;
}(CodeGeneratingComponentRendererCodeEmitter));
export { WPFXamlCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter, _super);
    function WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b) {
        return _super.call(this, a, b) || this;
    }
    WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.a9 = function (a) {
        if (a == null || isNaN_(a)) {
            return "";
        }
        return "TimeSpan.FromMilliseconds(" + a + ")";
    };
    WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_r = function () {
        return true;
    };
    Object.defineProperty(WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype, "r", {
        get: function () {
            return this.get_r();
        },
        enumerable: false,
        configurable: true
    });
    WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.i = function (a, b) {
        return new WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
    };
    WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter.prototype.get_y = function () {
        return false;
    };
    WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = markType(WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter;
}(CodeGeneratingComponentRendererCodeEmitter));
export { WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter };
/**
 * @hidden
 */
var CodeGeneratingComponentRenderer = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CodeGeneratingComponentRenderer, _super);
    function CodeGeneratingComponentRenderer(a, b) {
        var _this = _super.call(this) || this;
        _this.ex = null;
        _this.ey = 0;
        _this.e1 = null;
        _this.e2 = null;
        _this.e3 = null;
        _this.e0 = null;
        _this.ez = null;
        _this.ex = b;
        _this.ey = a;
        _this.preserveKeyOrder = true;
        ComponentRenderer.platform = CodeGeneratingComponentRenderer.e4(a);
        _this.adapter = new CodeGenerationRendererAdapter(a, _this.adapter);
        if (b.reusedContext != null) {
            _this.context = b.reusedContext;
        }
        else {
            _this.context = new TypeDescriptionContext(_this.adapter, ComponentRenderer.platform);
        }
        _this.k = new CodeGenerationTransformer(a);
        return _this;
    }
    CodeGeneratingComponentRenderer.prototype.loadCodeJson = function (a) {
        this.loadJson(a, function (b) {
            return ((function () {
                var $ret = new CodeGenerationContainer();
                $ret.id = b;
                return $ret;
            })());
        });
    };
    CodeGeneratingComponentRenderer.prototype.ds = function (a) {
        this.ex.c = a;
    };
    CodeGeneratingComponentRenderer.prototype.p = function () {
        return false;
    };
    CodeGeneratingComponentRenderer.e5 = function (a) {
        if (a == null) {
            return null;
        }
        if (a.length == 1) {
            return a.toLowerCase();
        }
        return a.substr(0, 1).toLowerCase() + a.substr(1);
    };
    CodeGeneratingComponentRenderer.e7 = function (a) {
        if (a == null) {
            return null;
        }
        if (a.length == 1) {
            return a.toLowerCase();
        }
        if (Base.equalsStatic(a.toUpperCase(), a)) {
            return a;
        }
        return a.substr(0, 1).toLowerCase() + a.substr(1);
    };
    CodeGeneratingComponentRenderer.e6 = function (a) {
        if (a == null) {
            return null;
        }
        if (a.length == 1) {
            return a.toUpperCase();
        }
        return a.substr(0, 1).toUpperCase() + a.substr(1);
    };
    CodeGeneratingComponentRenderer.e8 = function (a) {
        if (a == null) {
            return null;
        }
        var b = new List$1(String_$type, 0);
        var c = 0;
        for (var d = 0; d < a.length; d++) {
            var e = a.charAt(d);
            var f = e.toUpperCase();
            var g = e.toLowerCase();
            var h = (c == 0 && b.count > 0);
            var i = (c > 1);
            if (f == e) {
                if (h) {
                    b.add('-');
                }
                c++;
            }
            else if (g == e) {
                if (i) {
                    b.insert(b.count - 1, '-');
                }
                c = 0;
            }
            else {
                c = 0;
            }
            b.add(g);
        }
        var j = new StringBuilder(0);
        for (var k = 0; k < b.count; k++) {
            j.h(b._inner[k]);
        }
        return j.toString();
    };
    CodeGeneratingComponentRenderer.prototype.at = function () {
        return true;
    };
    CodeGeneratingComponentRenderer.prototype.bz = function (a) {
        var b = a.g;
        return b == null ? { $type: Point_$type, x: NaN, y: NaN } : { $type: Point_$type, x: b.x, y: b.y };
    };
    CodeGeneratingComponentRenderer.prototype.b0 = function (a) {
        var b = a.g;
        return b == null ? Rect.empty : new Rect(0, b.left, b.top, b.width, b.height);
    };
    CodeGeneratingComponentRenderer.prototype.b1 = function (a) {
        var b = a.g;
        return b == null ? new Size(1, NaN, NaN) : new Size(1, b.width, b.height);
    };
    CodeGeneratingComponentRenderer.prototype.b3 = function (a) {
        var b = a.g;
        if (typeof b === 'number') {
            return typeGetValue(b);
        }
        if (typeof b === 'number') {
            return b;
        }
        if (typeof b === 'string') {
            return parseFloat(b);
        }
        return NaN;
    };
    CodeGeneratingComponentRenderer.prototype.cu = function (a) {
        if (a == null) {
            return null;
        }
        return Brush.create(a);
    };
    CodeGeneratingComponentRenderer.prototype.cv = function (a) {
        if (a == null) {
            return null;
        }
        return Brush.create(a).color;
    };
    CodeGeneratingComponentRenderer.prototype.ag = function (a) {
        _super.prototype.ag.call(this, a);
        this.e1 = a;
        return true;
    };
    CodeGeneratingComponentRenderer.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        this.e3 = a;
        return true;
    };
    CodeGeneratingComponentRenderer.prototype.ah = function (a) {
        _super.prototype.ah.call(this, a);
        this.ez = a;
        return true;
    };
    CodeGeneratingComponentRenderer.prototype.ai = function (a) {
        _super.prototype.ai.call(this, a);
        this.e0 = a;
        return true;
    };
    CodeGeneratingComponentRenderer.prototype.aj = function (a) {
        _super.prototype.aj.call(this, a);
        this.e2 = a;
        return true;
    };
    CodeGeneratingComponentRenderer.prototype.er = function (a, b, c) {
        switch (this.ey) {
            case 1: if (b.forceCodeBehind) {
                return new AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
            }
            else {
                return new AngularTemplateCodeGeneratingComponentRendererCodeEmitter(b, c);
            }
            case 2: if (b.forceCodeBehind) {
                return new ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
            }
            else {
                return new ReactMarkupCodeGeneratingComponentRendererCodeEmitter(b, c);
            }
            case 3: if (b.forceCodeBehind) {
                return new WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
            }
            else {
                return new WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter(b, c);
            }
            case 4: break;
            case 0: if (b.forceCodeBehind) {
                return new WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
            }
            else {
                return new WPFXamlCodeGeneratingComponentRendererCodeEmitter(b, c);
            }
            case 5: if (b.forceCodeBehind) {
                return new BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
            }
            else {
                return new BlazorRazorCodeGeneratingComponentRendererCodeEmitter(b, c);
            }
        }
        return null;
    };
    CodeGeneratingComponentRenderer.prototype.es = function (a, b) {
        switch (this.ey) {
            case 1: return new TSCodeGeneratingComponentRendererDataEmitter(a, b);
            case 2: return new TSCodeGeneratingComponentRendererDataEmitter(a, b);
            case 3: return new TSCodeGeneratingComponentRendererDataEmitter(a, b);
            case 4: break;
            case 0:
            case 5: return new DotNetCodeGeneratingComponentRendererDataEmitter(a, b);
        }
        return null;
    };
    CodeGeneratingComponentRenderer.prototype.et = function (a, b) {
        switch (this.ey) {
            case 1: return new TSCodeGeneratingComponentRendererHandlerEmitter(a, b, this.ey);
            case 2: return new TSCodeGeneratingComponentRendererHandlerEmitter(a, b, this.ey);
            case 3: return new TSCodeGeneratingComponentRendererHandlerEmitter(a, b, this.ey);
            case 4: break;
            case 0:
            case 5: return new DotNetCodeGeneratingComponentRendererHandlerEmitter(a, b);
        }
        return null;
    };
    CodeGeneratingComponentRenderer.prototype.ev = function (a, b) {
        switch (this.ey) {
            case 1: return new TSCodeGeneratingComponentRendererTemplateEmitter(a, b, this.ey);
            case 2: return new TSCodeGeneratingComponentRendererTemplateEmitter(a, b, this.ey);
            case 3: return new TSCodeGeneratingComponentRendererTemplateEmitter(a, b, this.ey);
            case 4: break;
            case 0:
            case 5: return new DotNetCodeGeneratingComponentRendererTemplateEmitter(a, b);
        }
        return null;
    };
    CodeGeneratingComponentRenderer.prototype.eu = function (a, b) {
        switch (this.ey) {
            case 1: return new TSCodeGeneratingComponentRendererModuleEmitter(this.ey, a, b);
            case 2: return new TSCodeGeneratingComponentRendererModuleEmitter(this.ey, a, b);
            case 3: return new TSCodeGeneratingComponentRendererModuleEmitter(this.ey, a, b);
            case 4: break;
            case 0: return new DotNetCodeGeneratingComponentRendererModuleEmitter(this.ey, a, b);
            case 5: return new BlazorCodeGeneratingComponentRendererModuleEmitter(this.ey, a, b);
        }
        return null;
    };
    CodeGeneratingComponentRenderer.prototype.markRefUsed = function (a) {
        var e_11, _a;
        try {
            for (var _b = __values(fromEnum(this.bc.keys)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var b = _c.value;
                var c = b;
                c.markRefUsed(a);
            }
        }
        catch (e_11_1) {
            e_11 = { error: e_11_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_11)
                    throw e_11.error;
            }
        }
    };
    CodeGeneratingComponentRenderer.prototype.emitCode = function (a) {
        var e_12, _a, e_13, _b, e_14, _c;
        var b = new CodeGeneratingRendererResult();
        b.platform = this.ey;
        b.generationOptions = this.ex;
        try {
            for (var _d = __values(fromEnum(this.bc.keys)), _e = _d.next(); !_e.done; _e = _d.next()) {
                var c = _e.value;
                var d = c;
                var e = d.id;
                try {
                    for (var _f = (e_13 = void 0, __values(fromEnum(d.b))), _g = _f.next(); !_g.done; _g = _f.next()) {
                        var f = _g.value;
                        if (f.j("Name") != null) {
                            b.trackDescriptionField(e, f.j("Name"));
                        }
                        if (f.j("name") != null) {
                            b.trackDescriptionField(e, f.j("name"));
                        }
                    }
                }
                catch (e_13_1) {
                    e_13 = { error: e_13_1 };
                }
                finally {
                    try {
                        if (_g && !_g.done && (_b = _f.return))
                            _b.call(_f);
                    }
                    finally {
                        if (e_13)
                            throw e_13.error;
                    }
                }
            }
        }
        catch (e_12_1) {
            e_12 = { error: e_12_1 };
        }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return))
                    _a.call(_d);
            }
            finally {
                if (e_12)
                    throw e_12.error;
            }
        }
        try {
            for (var _h = __values(fromEnum(this.bc.keys)), _j = _h.next(); !_j.done; _j = _h.next()) {
                var g = _j.value;
                var h = g;
                h.d(this.context, b, this, this.ex);
            }
        }
        catch (e_14_1) {
            e_14 = { error: e_14_1 };
        }
        finally {
            try {
                if (_j && !_j.done && (_c = _h.return))
                    _c.call(_h);
            }
            finally {
                if (e_14)
                    throw e_14.error;
            }
        }
        if (this.e1 != null) {
            var i = this.eu(this.ex, b);
            i.n(this.e1, this.context);
        }
        if (this.e3 != null) {
            if (typeCast(JsonDictionaryObject.$, this.e3) !== null) {
                var j = this.e3;
                var m = j.e();
                for (var l = 0; l < m.length; l++) {
                    var k = m[l];
                    b.defineString(k, j.item(k).value);
                }
            }
        }
        {
            var n = this.es(this.ex, b);
            n.ac(this.e2);
        }
        var o = this.et(this.ex, b);
        var p = new List$1(String_$type, 0);
        var q = new List$1(String_$type, 0);
        if (this.ez != null) {
            if (typeCast(JsonDictionaryArray.$, this.ez) !== null) {
                var r = this.ez;
                if (r.items != null) {
                    var u = r.items;
                    for (var t = 0; t < u.length; t++) {
                        var s = u[t];
                        p.add(s.value.toString());
                    }
                }
            }
            else if (typeCast(JsonDictionaryValue.$, this.ez) !== null) {
                var v = this.ez;
                p.add(v.value.toString());
            }
        }
        if (this.e0 != null) {
            if (typeCast(JsonDictionaryArray.$, this.e0) !== null) {
                var w = this.e0;
                if (w.items != null) {
                    var z = w.items;
                    for (var y = 0; y < z.length; y++) {
                        var x = z[y];
                        q.add(x.value.toString());
                    }
                }
            }
            else if (typeCast(JsonDictionaryValue.$, this.e0) !== null) {
                var aa = this.e0;
                q.add(aa.value.toString());
            }
        }
        o.s(p.toArray(), q.toArray());
        var ab = this.ev(this.ex, b);
        ab.r();
        this.fd(b, a);
        return b;
    };
    CodeGeneratingComponentRenderer.prototype.fd = function (a, b) {
        b.execute(a);
    };
    CodeGeneratingComponentRenderer.prototype.bl = function () {
        return runOn(this, this.fa);
    };
    CodeGeneratingComponentRenderer.prototype.e9 = function (a) {
        a = stringReplace(a, "-", "_");
        if (!this.ex.g) {
            switch (this.ey) {
                case 1:
                case 2:
                case 3: return CodeGeneratingComponentRenderer.e5(a);
                case 4:
                case 0:
                case 5: return CodeGeneratingComponentRenderer.e6(a);
            }
        }
        return a;
    };
    CodeGeneratingComponentRenderer.prototype.fa = function (a) {
        var b = a;
        if (stringStartsWith(a, "{")) {
            if (!stringStartsWith(a, "{[")) {
                a = a.substr(1, a.length - 2);
                a = "{" + this.e9(a) + "}";
                b = a;
            }
        }
        else {
            b = this.e9(a);
        }
        return b;
    };
    CodeGeneratingComponentRenderer.e4 = function (a) {
        switch (a) {
            case 1: return 3;
            case 2: return 2;
            case 3: return 4;
            case 4: return 8;
            case 0: return 0;
            case 5: return 12;
        }
        return 0;
    };
    CodeGeneratingComponentRenderer.$t = markType(CodeGeneratingComponentRenderer, 'CodeGeneratingComponentRenderer', ComponentRenderer.$);
    return CodeGeneratingComponentRenderer;
}(ComponentRenderer));
export { CodeGeneratingComponentRenderer };
/**
 * @hidden
 */
var CodeGenerationContainer = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CodeGenerationContainer, _super);
    function CodeGenerationContainer() {
        var _this = _super.call(this) || this;
        _this._id = null;
        _this.b = null;
        _this.a = new HashSet$1(String_$type, 0);
        _this.b = new List$1(CodeGenerationItemBuilder.$, 0);
        return _this;
    }
    Object.defineProperty(CodeGenerationContainer.prototype, "id", {
        get: function () {
            return this._id;
        },
        set: function (a) {
            this._id = a;
        },
        enumerable: false,
        configurable: true
    });
    CodeGenerationContainer.prototype.d = function (a, b, c, d) {
        var e_15, _a, e_16, _b;
        var e = new StringBuilder(0);
        var f = new StringBuilder(0);
        var g = new StringBuilder(0);
        var h = new CodeGeneratingImportManager();
        var _loop_2 = function (i) {
            if (!c.hasRef(i)) {
                c.f(i);
            }
            b.af(((function () {
                var $ret = new CodeGenerationItemBuilderPropertyInfo();
                $ret.a = c.g(i);
                return $ret;
            })()));
        };
        try {
            for (var _c = __values(fromEnum(this.a)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var i = _d.value;
                _loop_2(i);
            }
        }
        catch (e_15_1) {
            e_15 = { error: e_15_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_15)
                    throw e_15.error;
            }
        }
        try {
            for (var _e = __values(fromEnum(this.b)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var j = _f.value;
                var k = c.er(j, d, b);
                j.v(k, true);
                var l = k.toString();
                b.addContainerResult(this.id, l);
                k.be(e);
                k.bh(f);
                k.bd(g);
                k.bf(a, h);
            }
        }
        catch (e_16_1) {
            e_16 = { error: e_16_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_16)
                    throw e_16.error;
            }
        }
        b.addContainerResult("bindingFields", e.toString());
        b.addContainerResult("bindingInit", f.toString());
        b.addContainerResult("bindingCode", g.toString());
        b.addImportResult("bindingImports", h);
    };
    CodeGenerationContainer.prototype.markRefUsed = function (a) {
        this.a.add_1(a);
    };
    CodeGenerationContainer.$t = markType(CodeGenerationContainer, 'CodeGenerationContainer');
    return CodeGenerationContainer;
}(Base));
export { CodeGenerationContainer };
/**
 * @hidden
 */
var CodeGenerationItemBuilder = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CodeGenerationItemBuilder, _super);
    function CodeGenerationItemBuilder(a, b, c, d) {
        var _this = _super.call(this) || this;
        _this.h = new List$1(String_$type, 0);
        _this.g = new Dictionary$2(String_$type, CodeGenerationItemBuilderPropertyInfo.$, 0);
        _this.c = null;
        _this.b = 0;
        _this.k = null;
        _this.i = null;
        _this.u = null;
        _this.b = a;
        _this.k = b;
        _this.c = c;
        _this.i = d;
        return _this;
    }
    CodeGenerationItemBuilder.prototype.p = function () {
        var a = this;
        var b = null;
        if (a.j("Name") != null) {
            b = a.j("Name");
        }
        if (a.j("name") != null) {
            b = a.j("name");
        }
        if (b != null) {
            return b;
        }
        return null;
    };
    CodeGenerationItemBuilder.prototype.a = function (a) {
        var b = this.c.getMetadata(this.k, a);
        return new CodeGenerationItemBuilder(this.b, this.s(b), this.c, this.i);
    };
    CodeGenerationItemBuilder.prototype.s = function (a) {
        return a.specificExternalType != null ? a.specificExternalType : a.specificType;
    };
    CodeGenerationItemBuilder.prototype.y = function (a, b, c, d) {
        if (!this.g.containsKey(a)) {
            this.h.add(a);
        }
        ;
        this.g.item(a, ((function () {
            var $ret = new CodeGenerationItemBuilderPropertyInfo();
            $ret.b = b;
            $ret.c = c;
            $ret.a = d;
            return $ret;
        })()));
    };
    CodeGenerationItemBuilder.prototype.w = function (a, b, c) {
        var d = null;
        if (!this.g.containsKey(a)) {
            this.h.add(a);
        }
        ;
        if (this.g.containsKey(a)) {
            var e = this.g.item(a);
            d = e.c;
        }
        this.g.item(a, ((function () {
            var $ret = new CodeGenerationItemBuilderPropertyInfo();
            $ret.b = b;
            $ret.c = d;
            $ret.a = c;
            return $ret;
        })()));
    };
    CodeGenerationItemBuilder.prototype.x = function (a, b) {
        if (this.g.containsKey(a)) {
            this.g.removeItem(a);
        }
    };
    CodeGenerationItemBuilder.prototype.j = function (a) {
        if (this.g.containsKey(a)) {
            return this.g.item(a).c;
        }
        return null;
    };
    CodeGenerationItemBuilder.prototype.v = function (a, b) {
        var e_17, _a, e_18, _b;
        var c = this.n();
        if (b) {
            var d = this.q();
            a.a5 = d;
        }
        var e = false;
        var f = false;
        if (this.g.containsKey("Name")) {
            e = true;
            a.bp(this.g.item("Name").c);
        }
        else if (this.g.containsKey("name")) {
            e = true;
            a.bp(this.g.item("name").c);
        }
        else {
            if (b && this.u != null) {
                e = true;
                a.bp(this.u);
            }
            else if (this.f(this.c, a)) {
                f = true;
                a.bo(c);
                if (a.y) {
                    this.h.add("name");
                    this.g.addItem("name", ((function () {
                        var $ret = new CodeGenerationItemBuilderPropertyInfo();
                        $ret.c = a.ap;
                        return $ret;
                    })()));
                }
                else {
                    this.h.add("Name");
                    this.g.addItem("Name", ((function () {
                        var $ret = new CodeGenerationItemBuilderPropertyInfo();
                        $ret.c = a.ap;
                        return $ret;
                    })()));
                }
            }
        }
        a.bc(this.r(), this.k, this.c);
        var g = 0;
        try {
            for (var _c = __values(fromEnum(this.h)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var h = _d.value;
                var i = this.g.item(h);
                if (i.b == null && h.toLowerCase() == "type") {
                    continue;
                }
                if (this.e(a, h, i)) {
                    a.bi(g, h, i, this.c);
                    g++;
                }
            }
        }
        catch (e_17_1) {
            e_17 = { error: e_17_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_17)
                    throw e_17.error;
            }
        }
        a.bq();
        var j = 0;
        try {
            for (var _e = __values(fromEnum(this.h)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var k = _f.value;
                var l = this.g.item(k);
                if (l.b == null && k.toLowerCase() == "type") {
                    continue;
                }
                if (!this.e(a, k, l)) {
                    a.bj(j, k, l, this.c);
                    j++;
                }
            }
        }
        catch (e_18_1) {
            e_18 = { error: e_18_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_18)
                    throw e_18.error;
            }
        }
        a.bk(j);
        a.bl(c);
        if (e) {
            a.bn();
        }
        else {
            if (f) {
                a.bm();
            }
        }
    };
    CodeGenerationItemBuilder.prototype.f = function (a, b) {
        var e_19, _a;
        if (b.n) {
            return true;
        }
        try {
            for (var _b = __values(fromEnum(this.h)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                var d = this.g.item(c);
                if (b.q(a, d)) {
                    return true;
                }
            }
        }
        catch (e_19_1) {
            e_19 = { error: e_19_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_19)
                    throw e_19.error;
            }
        }
        return false;
    };
    CodeGenerationItemBuilder.prototype.d = function (a) {
        var e_20, _a;
        var b = typeCast(Array_$type, a.c);
        try {
            for (var b_1 = __values(b), b_1_1 = b_1.next(); !b_1_1.done; b_1_1 = b_1.next()) {
                var c = b_1_1.value;
                if (typeCast(CodeGenerationItemBuilder.$, c) !== null) {
                    return false;
                }
            }
        }
        catch (e_20_1) {
            e_20 = { error: e_20_1 };
        }
        finally {
            try {
                if (b_1_1 && !b_1_1.done && (_a = b_1.return))
                    _a.call(b_1);
            }
            finally {
                if (e_20)
                    throw e_20.error;
            }
        }
        return true;
    };
    CodeGenerationItemBuilder.prototype.e = function (a, b, c) {
        if (c.b == null) {
            if (b.toLowerCase() == "name") {
                return true;
            }
        }
        switch (c.b.knownType) {
            case 10: return a.r ? this.d(c) : true;
            case 7: return true;
            case 4: return c.c == null || !(typeCast(LinearGradientBrush.$, c.c) !== null);
            case 6:
                return true;
                break;
            case 23: return true;
            case 9:
                if (a.w(this.c, c)) {
                    return true;
                }
                return false;
            case 5: return true;
            case 16: return true;
            case 20: return true;
            case 22: return true;
            case 15: return false;
            case 3: return true;
            case 19: return true;
            case 8:
                if (a.w(this.c, c)) {
                    return true;
                }
                if (c.b.specificExternalType == "string") {
                    return true;
                }
                if (c.a != null) {
                    return true;
                }
                return false;
            case 13: return false;
            case 18: return true;
            case 1: return true;
            case 11: return true;
            case 14: return true;
            case 12: return true;
            case 2: return true;
            case 21: return true;
            case 17: return true;
            case 0: return false;
        }
        return false;
    };
    CodeGenerationItemBuilder.prototype.r = function () {
        return this.n();
    };
    Object.defineProperty(CodeGenerationItemBuilder.prototype, "t", {
        get: function () {
            return this.k;
        },
        enumerable: false,
        configurable: true
    });
    CodeGenerationItemBuilder.o = function (a, b, c, d, e) {
        if (a == 0 || a == 4) {
            var f = c.getMetadata(b, "_qualifiedName");
            if (f != null) {
                var g = f.specificExternalType;
                var h = g.lastIndexOf('.');
                if (h >= 0) {
                    return g.substr(h + 1);
                }
                else {
                    return g;
                }
            }
            return b;
        }
        else if ((a == 5 || a == 2 || a == 1 || a == 3) && c.getMetadata(b, "__qualifiedNameTS") != null) {
            var i = c.getMetadata(b, "__qualifiedNameTS");
            if (c.getMetadata(b, "__skipSuffix") != null) {
                return d + i.specificExternalType;
            }
            return d + i.specificExternalType + e;
        }
        else {
            if (c.getMetadata(b, "__skipSuffix") != null) {
                return d + b;
            }
            return d + b + e;
        }
    };
    CodeGenerationItemBuilder.prototype.n = function () {
        return CodeGenerationItemBuilder.o(this.b, this.k, this.c, this.l, this.m);
    };
    CodeGenerationItemBuilder.prototype.q = function () {
        var a = this.c.getMetadata(this.k, "_qualifiedName");
        if (a != null) {
            var b = a.specificExternalType;
            var c = b.split('.');
            if (c.length > 2) {
                return c[c.length - 2].toLowerCase();
            }
        }
        if (this.b == 1 || this.b == 3 || this.b == 2 || this.b == 5) {
            a = this.c.getMetadata(this.k, "__packageNameWC");
            if (a != null) {
                var d = a.specificExternalType;
                return d;
            }
        }
        if (this.k == "Legend" || this.k == "ItemLegend") {
            return "charts";
        }
        return "core";
    };
    CodeGenerationItemBuilder.prototype.get_l = function () {
        switch (this.b) {
            case 1: return "Igx";
            case 2: return "Igr";
            case 3: return "Igc";
            case 4: return "Ultra";
            case 0: return "Xam";
            case 5: return "Igb";
        }
        return "";
    };
    Object.defineProperty(CodeGenerationItemBuilder.prototype, "l", {
        get: function () {
            return this.get_l();
        },
        enumerable: false,
        configurable: true
    });
    CodeGenerationItemBuilder.prototype.get_m = function () {
        switch (this.b) {
            case 1: return "Component";
            case 2: return "";
            case 3: return "Component";
            case 4: return "";
            case 0: return "";
        }
        return "";
    };
    Object.defineProperty(CodeGenerationItemBuilder.prototype, "m", {
        get: function () {
            return this.get_m();
        },
        enumerable: false,
        configurable: true
    });
    CodeGenerationItemBuilder.$t = markType(CodeGenerationItemBuilder, 'CodeGenerationItemBuilder');
    return CodeGenerationItemBuilder;
}(Base));
export { CodeGenerationItemBuilder };
/**
 * @hidden
 */
var CodeGenerationRendererAdapter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CodeGenerationRendererAdapter, _super);
    function CodeGenerationRendererAdapter(a, b) {
        var _this = _super.call(this) || this;
        _this.a = 0;
        _this.b = null;
        _this.b = b;
        _this.a = a;
        return _this;
    }
    CodeGenerationRendererAdapter.prototype.createColorCollection = function (a) {
        var b = new List$1(Base.$, 0);
        for (var c = 0; c < a.length; c++) {
            b.add1(a[c]);
        }
        return b;
    };
    CodeGenerationRendererAdapter.prototype.createBrushCollection = function (a) {
        var b = new List$1(Base.$, 0);
        for (var c = 0; c < a.length; c++) {
            b.add1(a[c]);
        }
        return b;
    };
    CodeGenerationRendererAdapter.prototype.createDoubleCollection = function (a) {
        var b = new List$1(Base.$, 0);
        for (var c = 0; c < a.length; c++) {
            b.add1(a[c]);
        }
        return b.toArray();
    };
    CodeGenerationRendererAdapter.prototype.createObject = function (a, b, c, d) {
        var e = new CodeGenerationItemBuilder(this.a, a, c, b);
        return e;
    };
    CodeGenerationRendererAdapter.prototype.coerceToEnum = function (a, b, c) {
        return c;
    };
    CodeGenerationRendererAdapter.prototype.onUIThread = function (a, b) {
        b();
    };
    CodeGenerationRendererAdapter.prototype.setOrUpdateCollectionOnTarget = function (a, b, c, d, e, f) {
        var g = this.getPropertyValue(e, b);
        if (g == null) {
            var h = new List$1(Base.$, 0);
            var i = f;
            for (var j = 0; j < i.count; j++) {
                h.add1(i.item(j));
            }
            this.setPropertyValue(e, b, c, h, null, null);
        }
        else {
            var k = f;
            var l = g;
            l.clear();
            for (var m = 0; m < k.count; m++) {
                l.add(k.item(m));
            }
        }
    };
    CodeGenerationRendererAdapter.prototype.c = function (a) {
        if (a == null) {
            return false;
        }
        return false;
    };
    CodeGenerationRendererAdapter.prototype.onPendingRef = function (a, b, c, d) {
        var e = a;
        e.w(b, c, d);
    };
    CodeGenerationRendererAdapter.prototype.setPropertyValue = function (a, b, c, d, e, f) {
        if (!(typeCast(CodeGenerationItemBuilder.$, a) !== null)) {
            this.b.setPropertyValue(a, b, c, d, e, f);
            return;
        }
        while (stringContains(b, ".")) {
            var g = b.split('.');
            var h = this.getPropertyValue(a, g[0]);
            if (h != null && !this.c(h)) {
                b = "";
                for (var i = 1; i < g.length; i++) {
                    if (i > 1) {
                        b += ".";
                    }
                    b += g[i];
                }
                a = h;
            }
            else {
                var j = a;
                var k = j.a(g[0]);
                h = k;
                this.setPropertyValue(a, g[0], c, h, null, f);
                b = "";
                for (var l = 1; l < g.length; l++) {
                    if (l > 1) {
                        b += ".";
                    }
                    b += g[l];
                }
                a = h;
            }
        }
        var m = a;
        m.y(b, c, d, f);
    };
    CodeGenerationRendererAdapter.prototype.getPropertyValue = function (a, b) {
        while (stringContains(b, ".")) {
            var c = b.split('.');
            var d = this.getPropertyValue(a, c[0]);
            if (d != null) {
                b = "";
                for (var e = 1; e < c.length; e++) {
                    if (e > 1) {
                        b += ".";
                    }
                    b += c[e];
                }
                a = d;
            }
            else {
                var f = a;
                var g = f.a(c[0]);
                d = g;
                for (var h = 1; h < c.length; h++) {
                    if (h > 1) {
                        b += ".";
                    }
                    b += c[h];
                }
                a = d;
            }
        }
        var i = a;
        return i.j(b);
    };
    CodeGenerationRendererAdapter.prototype.getRootObject = function (a) {
        var b = a;
        var c = b.b.count > 0 ? b.b._inner[0] : null;
        return c;
    };
    CodeGenerationRendererAdapter.prototype.clearContainer = function (a, b, c) {
        var d = a;
        var e = d.b.count > 0 ? d.b._inner[0] : null;
        d.b.clear();
        c(false);
    };
    CodeGenerationRendererAdapter.prototype.forPropertyValueItem = function (a, b, c) {
        var d = this.getPropertyValue(a, b);
        if (typeCast(IList_$type, d) !== null) {
            var e = d;
            for (var f = 0; f < e.count; f++) {
                c(e.item(f));
            }
        }
        else if (typeCast(Array_$type, d) !== null) {
            var g = d;
            for (var h = 0; h < g.length; h++) {
                c(g[h]);
            }
        }
    };
    CodeGenerationRendererAdapter.prototype.clearCollection = function (a, b, c) {
        var d = this.getPropertyValue(a, b);
        if (typeCast(IList_$type, d) !== null) {
            d.clear();
        }
        else if (typeCast(Array_$type, d) !== null) {
            this.setPropertyValue(a, b, c, null, d, null);
        }
    };
    CodeGenerationRendererAdapter.prototype.addItemToCollection = function (a, b, c, d, e) {
        var f = this.getPropertyValue(c, a);
        if (typeCast(IList_$type, f) !== null) {
            f.insert(d, e);
        }
        else if (typeCast(Array_$type, f) !== null) {
            var g = f;
            var h = new List$1(Base.$, 0);
            for (var i = 0; i < g.count; i++) {
                h.add1(g.item(i));
            }
            h.insert1(d, e);
            var j = [g.count + 1];
            for (var k = 0; k < j.length; k++) {
                j[k] = h._inner[k];
            }
            this.setPropertyValue(c, a, b, j, f, null);
        }
    };
    CodeGenerationRendererAdapter.prototype.resetPropertyOnTarget = function (a, b, c, d) {
        while (stringContains(b, ".")) {
            var e = b.split('.');
            var f = this.getPropertyValue(d, e[0]);
            if (f != null && !this.c(f)) {
                b = "";
                for (var g = 1; g < e.length; g++) {
                    if (g > 1) {
                        b += ".";
                    }
                    b += e[g];
                }
                d = f;
            }
            else {
                var h = d;
                var i = h.a(e[0]);
                f = i;
                this.setPropertyValue(d, e[0], c, f, null, null);
                b = "";
                for (var j = 1; j < e.length; j++) {
                    if (j > 1) {
                        b += ".";
                    }
                    b += e[j];
                }
                d = f;
            }
        }
        var k = d;
        k.x(b, c);
    };
    CodeGenerationRendererAdapter.prototype.replaceItemInCollection = function (a, b, c, d, e) {
        var f = this.getPropertyValue(c, a);
        if (typeCast(IList_$type, f) !== null) {
            f.item(d, e);
        }
        else if (typeCast(Array_$type, f) !== null) {
            var g = f;
            var h = new List$1(Base.$, 0);
            for (var i = 0; i < g.count; i++) {
                h.add1(g.item(i));
            }
            h._inner[d] = e;
            var j = [g.count + 1];
            for (var k = 0; k < j.length; k++) {
                j[k] = h._inner[k];
            }
            this.setPropertyValue(c, a, b, j, f, null);
        }
    };
    CodeGenerationRendererAdapter.prototype.replaceRootItem = function (a, b, c, d) {
        var e = this.createObject(b, a, c, null);
        var f = a;
        var g = f.b.count > 0 ? f.b._inner[0] : null;
        f.b.clear();
        {
            f.b.add(e);
        }
        d(false);
    };
    CodeGenerationRendererAdapter.prototype.removeItemFromCollection = function (a, b, c, d) {
        var e = this.getPropertyValue(c, a);
        if (typeCast(IList_$type, e) !== null) {
            e.removeAt(d);
        }
        else if (typeCast(Array_$type, e) !== null) {
            var f = e;
            var g = new List$1(Base.$, 0);
            for (var h = 0; h < f.count; h++) {
                g.add1(f.item(h));
            }
            g.removeAt(d);
            var i = [f.count - 1];
            for (var j = 0; j < i.length; j++) {
                i[j] = g._inner[j];
            }
            this.setPropertyValue(c, a, b, i, e, null);
        }
    };
    CodeGenerationRendererAdapter.prototype.removeRootItem = function (a, b, c) {
        var d = a;
        var e = d.b.count > 0 ? d.b._inner[0] : null;
        d.b.clear();
        c(false);
    };
    CodeGenerationRendererAdapter.prototype.flushChanges = function (a) {
    };
    CodeGenerationRendererAdapter.prototype.executeMethod = function (a, b, c, d, e) {
    };
    CodeGenerationRendererAdapter.prototype.serializeBrush = function (a) {
        return null;
    };
    CodeGenerationRendererAdapter.prototype.serializeColor = function (a) {
        return null;
    };
    CodeGenerationRendererAdapter.prototype.serializeBrushCollection = function (a) {
        return null;
    };
    CodeGenerationRendererAdapter.prototype.serializePoint = function (a) {
        return null;
    };
    CodeGenerationRendererAdapter.prototype.serializeSize = function (a) {
        return null;
    };
    CodeGenerationRendererAdapter.prototype.serializeRect = function (a) {
        return null;
    };
    CodeGenerationRendererAdapter.prototype.serializeColorCollection = function (a) {
        return null;
    };
    CodeGenerationRendererAdapter.prototype.serializeTimespan = function (a) {
        return null;
    };
    CodeGenerationRendererAdapter.prototype.serializeDoubleCollection = function (a) {
        return null;
    };
    CodeGenerationRendererAdapter.$t = markType(CodeGenerationRendererAdapter, 'CodeGenerationRendererAdapter', Base.$, [IComponentRendererAdapter_$type]);
    return CodeGenerationRendererAdapter;
}(Base));
export { CodeGenerationRendererAdapter };
