// igniteui-react-core/esm5/lib/ColorEditorPanelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ColorEditorPanelDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ColorEditorPanelDescription, _super);
    function ColorEditorPanelDescription() {
        var _this = _super.call(this) || this;
        _this.y = null;
        _this.p = null;
        _this.o = null;
        _this.n = null;
        _this.v = null;
        _this.w = null;
        _this.q = null;
        _this.u = null;
        _this.x = null;
        _this.r = null;
        _this.i = 0;
        _this.s = null;
        _this.t = null;
        _this.j = 0;
        _this.h = 0;
        _this.z = null;
        _this.aa = null;
        return _this;
    }
    ColorEditorPanelDescription.prototype.get_type = function () {
        return "ColorEditorPanel";
    };
    Object.defineProperty(ColorEditorPanelDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "value", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "density", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Density");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "baseTheme", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "backgroundColor", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("BackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "selectedColorBorderColor", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("SelectedColorBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "selectedFocusDateBackgroundColor", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("SelectedFocusDateBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "focusColorBorderColor", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("FocusColorBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "hoverBackgroundColor", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("HoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "textColor", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "fontFamily", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("FontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "fontSize", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("FontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "fontStyle", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("FontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "fontWeight", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("FontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("PixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "actualPixelScalingRatio", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ActualPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "valueChangedRef", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("ValueChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColorEditorPanelDescription.prototype, "valueChangingRef", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("ValueChangingRef");
        },
        enumerable: false,
        configurable: true
    });
    ColorEditorPanelDescription.$t = markType(ColorEditorPanelDescription, 'ColorEditorPanelDescription', Description.$);
    return ColorEditorPanelDescription;
}(Description));
export { ColorEditorPanelDescription };
