// igniteui-react-core/esm5/lib/ColumnSummaryDescriptionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ColumnSummaryDescriptionDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ColumnSummaryDescriptionDescription, _super);
    function ColumnSummaryDescriptionDescription() {
        var _this = _super.call(this) || this;
        _this.u = null;
        _this.w = null;
        _this.r = null;
        _this.t = null;
        _this.y = null;
        _this.m = false;
        _this.i = null;
        _this.s = null;
        _this.h = null;
        _this.o = 0;
        _this.v = null;
        _this.x = null;
        return _this;
    }
    ColumnSummaryDescriptionDescription.prototype.get_type = function () {
        return "ColumnSummaryDescription";
    };
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "field", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Field");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "operand", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("Operand");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "calculatorDisplayName", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("CalculatorDisplayName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "displayName", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("DisplayName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "valueFormat", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("ValueFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "shouldDisplay", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("ShouldDisplay");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "valueFormatSpecifiers", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ValueFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "displayFormat", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("DisplayFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "displayFormatSpecifiers", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("DisplayFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "maxFractionDigits", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("MaxFractionDigits");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "formatTextRef", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("FormatTextRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnSummaryDescriptionDescription.prototype, "provideCalculatorRef", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("ProvideCalculatorRef");
        },
        enumerable: false,
        configurable: true
    });
    ColumnSummaryDescriptionDescription.$t = markType(ColumnSummaryDescriptionDescription, 'ColumnSummaryDescriptionDescription', Description.$);
    ColumnSummaryDescriptionDescription.__marshalByValue = true;
    ColumnSummaryDescriptionDescription.__marshalByValueAlias = "ColumnSummaryDescription";
    return ColumnSummaryDescriptionDescription;
}(Description));
export { ColumnSummaryDescriptionDescription };
