// igniteui-react-core/esm5/lib/ColumnGroupDescriptionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ColumnSortDescriptionDescription } from "./ColumnSortDescriptionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ColumnGroupDescriptionDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ColumnGroupDescriptionDescription, _super);
    function ColumnGroupDescriptionDescription() {
        var _this = _super.call(this) || this;
        _this.x = null;
        _this.p = null;
        _this.u = null;
        _this.o = null;
        _this.v = null;
        _this.w = null;
        return _this;
    }
    ColumnGroupDescriptionDescription.prototype.get_type = function () {
        return "ColumnGroupDescription";
    };
    Object.defineProperty(ColumnGroupDescriptionDescription.prototype, "valueFormat", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("ValueFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupDescriptionDescription.prototype, "valueFormatSpecifiers", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("ValueFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupDescriptionDescription.prototype, "displayFormat", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("DisplayFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupDescriptionDescription.prototype, "displayFormatSpecifiers", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("DisplayFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupDescriptionDescription.prototype, "displayName", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("DisplayName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupDescriptionDescription.prototype, "formatTextRef", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("FormatTextRef");
        },
        enumerable: false,
        configurable: true
    });
    ColumnGroupDescriptionDescription.$t = markType(ColumnGroupDescriptionDescription, 'ColumnGroupDescriptionDescription', ColumnSortDescriptionDescription.$);
    ColumnGroupDescriptionDescription.__marshalByValue1 = true;
    ColumnGroupDescriptionDescription.__marshalByValueAlias1 = "ColumnGroupDescription";
    return ColumnGroupDescriptionDescription;
}(ColumnSortDescriptionDescription));
export { ColumnGroupDescriptionDescription };
