// igniteui-react-core/esm5/lib/ColumnGroupingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ColumnGroupingDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ColumnGroupingDescription, _super);
    function ColumnGroupingDescription() {
        var _this = _super.call(this) || this;
        _this.s = null;
        _this.k = null;
        _this.o = null;
        _this.u = null;
        _this.v = null;
        _this.l = null;
        _this.h = 0;
        _this.m = null;
        _this.n = null;
        _this.t = null;
        _this.p = null;
        _this.q = null;
        _this.r = null;
        _this.j = null;
        return _this;
    }
    ColumnGroupingDescription.prototype.get_type = function () {
        return "ColumnGrouping";
    };
    Object.defineProperty(ColumnGroupingDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "targetGridRef", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("TargetGridRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "baseTheme", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "density", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Density");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "title", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "titleColor", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("TitleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "columnGroupTitleFontFamily", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ColumnGroupTitleFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "columnGroupTitleFontSize", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ColumnGroupTitleFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "columnGroupTitleFontStyle", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("ColumnGroupTitleFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "columnGroupTitleFontWeight", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ColumnGroupTitleFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "textColor", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "iconColor", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("IconColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "itemBackgroundColor", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("ItemBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "itemHoverBackgroundColor", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("ItemHoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ColumnGroupingDescription.prototype, "backgroundColor", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("BackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    ColumnGroupingDescription.$t = markType(ColumnGroupingDescription, 'ColumnGroupingDescription', Description.$);
    return ColumnGroupingDescription;
}(Description));
export { ColumnGroupingDescription };
