// igniteui-react-core/esm5/lib/ComboEditorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ComboEditorDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ComboEditorDescription, _super);
    function ComboEditorDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.cx = null;
        _this.cj = null;
        _this.cy = null;
        _this.ck = null;
        _this.cz = null;
        _this.cl = null;
        _this.at = 0;
        _this.ab = 0;
        _this.au = 0;
        _this.av = 0;
        _this.aw = 0;
        _this.ax = 0;
        _this.ac = 0;
        _this.ad = 0;
        _this.ae = 0;
        _this.af = 0;
        _this.az = 0;
        _this.ay = 0;
        _this.a0 = 0;
        _this.a1 = 0;
        _this.ah = 0;
        _this.ag = 0;
        _this.ai = 0;
        _this.aj = 0;
        _this.i = null;
        _this.c3 = null;
        _this.cm = null;
        _this.r = false;
        _this.c4 = null;
        _this.cn = null;
        _this.a2 = 0;
        _this.ak = 0;
        _this.c5 = null;
        _this.co = null;
        _this.a3 = 0;
        _this.al = 0;
        _this.a4 = 0;
        _this.am = 0;
        _this.da = null;
        _this.cp = null;
        _this.a6 = 0;
        _this.an = 0;
        _this.a7 = 0;
        _this.ao = 0;
        _this.c2 = null;
        _this.dc = null;
        _this.dg = null;
        _this.cq = null;
        _this.dd = null;
        _this.a8 = 0;
        _this.de = null;
        _this.df = null;
        _this.p = false;
        _this.di = null;
        _this.cr = null;
        _this.dj = null;
        _this.cs = null;
        _this.dn = null;
        _this.ct = null;
        _this.dk = null;
        _this.a9 = 0;
        _this.dl = null;
        _this.dm = null;
        _this.dp = null;
        _this.dt = null;
        _this.cu = null;
        _this.c6 = null;
        _this.a5 = 0;
        _this.c7 = null;
        _this.c8 = null;
        _this.dw = null;
        _this.cv = null;
        _this.ba = 0;
        _this.ap = 0;
        _this.dx = null;
        _this.cw = null;
        _this.bb = 0;
        _this.aq = 0;
        _this.bc = 0;
        _this.ar = 0;
        _this.bd = 0;
        _this.as = 0;
        _this.q = false;
        _this.du = null;
        _this.dr = null;
        _this.k = null;
        _this.h = null;
        _this.ch = null;
        _this.s = false;
        _this.t = false;
        _this.u = false;
        _this.c9 = null;
        _this.dh = null;
        _this.dv = null;
        _this.dq = null;
        _this.db = null;
        _this.c0 = null;
        _this.c1 = null;
        _this.ds = null;
        _this.dy = null;
        return _this;
    }
    ComboEditorDescription.prototype.get_type = function () {
        return "ComboEditor";
    };
    Object.defineProperty(ComboEditorDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "fields", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Fields");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "backgroundColor", {
        get: function () {
            return this.cx;
        },
        set: function (a) {
            this.cx = a;
            this.g("BackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualBackgroundColor", {
        get: function () {
            return this.cj;
        },
        set: function (a) {
            this.cj = a;
            this.g("ActualBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "baseTheme", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualBaseTheme", {
        get: function () {
            return this.ck;
        },
        set: function (a) {
            this.ck = a;
            this.g("ActualBaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "borderColor", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("BorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualBorderColor", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            this.cl = a;
            this.g("ActualBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "borderWidth", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("BorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualBorderWidth", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("ActualBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "contentPaddingBottom", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("ContentPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "contentPaddingLeft", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("ContentPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "contentPaddingRight", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("ContentPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "contentPaddingTop", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("ContentPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualContentPaddingBottom", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("ActualContentPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualContentPaddingLeft", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("ActualContentPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualContentPaddingRight", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("ActualContentPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualContentPaddingTop", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("ActualContentPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "cornerRadiusBottomRight", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("CornerRadiusBottomRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "cornerRadiusBottomLeft", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("CornerRadiusBottomLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "cornerRadiusTopLeft", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("CornerRadiusTopLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "cornerRadiusTopRight", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("CornerRadiusTopRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualCornerRadiusBottomRight", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("ActualCornerRadiusBottomRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualCornerRadiusBottomLeft", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("ActualCornerRadiusBottomLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualCornerRadiusTopLeft", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("ActualCornerRadiusTopLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualCornerRadiusTopRight", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("ActualCornerRadiusTopRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "dataSourceDesiredProperties", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("DataSourceDesiredProperties");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "density", {
        get: function () {
            return this.c3;
        },
        set: function (a) {
            this.c3 = a;
            this.g("Density");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualDensity", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.g("ActualDensity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "dropDownButtonVisible", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("DropDownButtonVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "focusBorderColor", {
        get: function () {
            return this.c4;
        },
        set: function (a) {
            this.c4 = a;
            this.g("FocusBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualFocusBorderColor", {
        get: function () {
            return this.cn;
        },
        set: function (a) {
            this.cn = a;
            this.g("ActualFocusBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "focusBorderWidth", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("FocusBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualFocusBorderWidth", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("ActualFocusBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "focusUnderlineColor", {
        get: function () {
            return this.c5;
        },
        set: function (a) {
            this.c5 = a;
            this.g("FocusUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualFocusUnderlineColor", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.g("ActualFocusUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "focusUnderlineOpacity", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("FocusUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualFocusUnderlineOpacity", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("ActualFocusUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "focusUnderlineRippleOpacity", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("FocusUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualFocusUnderlineRippleOpacity", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("ActualFocusUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "hoverUnderlineColor", {
        get: function () {
            return this.da;
        },
        set: function (a) {
            this.da = a;
            this.g("HoverUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualHoverUnderlineColor", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("ActualHoverUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "hoverUnderlineOpacity", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("HoverUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualHoverUnderlineOpacity", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("ActualHoverUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "hoverUnderlineWidth", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("HoverUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualHoverUnderlineWidth", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("ActualHoverUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.c2;
        },
        set: function (a) {
            this.c2 = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "label", {
        get: function () {
            return this.dc;
        },
        set: function (a) {
            this.dc = a;
            this.g("Label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "labelTextColor", {
        get: function () {
            return this.dg;
        },
        set: function (a) {
            this.dg = a;
            this.g("LabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualLabelTextColor", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("ActualLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "labelFontFamily", {
        get: function () {
            return this.dd;
        },
        set: function (a) {
            this.dd = a;
            this.g("LabelFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "labelFontSize", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("LabelFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "labelFontStyle", {
        get: function () {
            return this.de;
        },
        set: function (a) {
            this.de = a;
            this.g("LabelFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "labelFontWeight", {
        get: function () {
            return this.df;
        },
        set: function (a) {
            this.df = a;
            this.g("LabelFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualLabelVisible", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("ActualLabelVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "noMatchesFoundLabel", {
        get: function () {
            return this.di;
        },
        set: function (a) {
            this.di = a;
            this.g("NoMatchesFoundLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualNoMatchesFoundLabel", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("ActualNoMatchesFoundLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "noMatchesFoundLabelBackgroundColor", {
        get: function () {
            return this.dj;
        },
        set: function (a) {
            this.dj = a;
            this.g("NoMatchesFoundLabelBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualNoMatchesFoundLabelBackgroundColor", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("ActualNoMatchesFoundLabelBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "noMatchesFoundLabelTextColor", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.g("NoMatchesFoundLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualNoMatchesFoundLabelTextColor", {
        get: function () {
            return this.ct;
        },
        set: function (a) {
            this.ct = a;
            this.g("ActualNoMatchesFoundLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "noMatchesFoundLabelFontFamily", {
        get: function () {
            return this.dk;
        },
        set: function (a) {
            this.dk = a;
            this.g("NoMatchesFoundLabelFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "noMatchesFoundLabelFontSize", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("NoMatchesFoundLabelFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "noMatchesFoundLabelFontStyle", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            this.dl = a;
            this.g("NoMatchesFoundLabelFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "noMatchesFoundLabelFontWeight", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            this.dm = a;
            this.g("NoMatchesFoundLabelFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "placeholder", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("Placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "textColor", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualTextColor", {
        get: function () {
            return this.cu;
        },
        set: function (a) {
            this.cu = a;
            this.g("ActualTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "fontFamily", {
        get: function () {
            return this.c6;
        },
        set: function (a) {
            this.c6 = a;
            this.g("FontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "fontSize", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("FontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "fontStyle", {
        get: function () {
            return this.c7;
        },
        set: function (a) {
            this.c7 = a;
            this.g("FontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "fontWeight", {
        get: function () {
            return this.c8;
        },
        set: function (a) {
            this.c8 = a;
            this.g("FontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "underlineColor", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.g("UnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualUnderlineColor", {
        get: function () {
            return this.cv;
        },
        set: function (a) {
            this.cv = a;
            this.g("ActualUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "underlineOpacity", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("UnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualUnderlineOpacity", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("ActualUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "underlineRippleColor", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("UnderlineRippleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualUnderlineRippleColor", {
        get: function () {
            return this.cw;
        },
        set: function (a) {
            this.cw = a;
            this.g("ActualUnderlineRippleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "underlineRippleOpacity", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("UnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualUnderlineRippleOpacity", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("ActualUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "underlineRippleWidth", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("UnderlineRippleWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualUnderlineRippleWidth", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("ActualUnderlineRippleWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "underlineWidth", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("UnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualUnderlineWidth", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("ActualUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "allowFilter", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("AllowFilter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "textField", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.g("TextField");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "text", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("Text");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "valueField", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("ValueField");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "actualValueField", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ActualValueField");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "value", {
        get: function () {
            return this.ch;
        },
        set: function (a) {
            this.ch = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "isFixed", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("IsFixed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "openAsChild", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("OpenAsChild");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "useTopLayer", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("UseTopLayer");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "gotFocusRef", {
        get: function () {
            return this.c9;
        },
        set: function (a) {
            this.c9 = a;
            this.g("GotFocusRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "lostFocusRef", {
        get: function () {
            return this.dh;
        },
        set: function (a) {
            this.dh = a;
            this.g("LostFocusRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "textValueChangedRef", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.g("TextValueChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "selectedValueChangedRef", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("SelectedValueChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "keyDownRef", {
        get: function () {
            return this.db;
        },
        set: function (a) {
            this.db = a;
            this.g("KeyDownRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "changeRef", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "changingRef", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.g("ChangingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "textChangeRef", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("TextChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboEditorDescription.prototype, "valueChangeRef", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("ValueChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    ComboEditorDescription.$t = markType(ComboEditorDescription, 'ComboEditorDescription', Description.$);
    return ComboEditorDescription;
}(Description));
export { ComboEditorDescription };
