// igniteui-react-core/esm5/lib/ComboBoxColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DataGridColumnDescription } from "./DataGridColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ComboBoxColumnDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ComboBoxColumnDescription, _super);
    function ComboBoxColumnDescription() {
        var _this = _super.call(this) || this;
        _this.f5 = null;
        _this.f1 = null;
        _this.f6 = null;
        _this.f7 = null;
        return _this;
    }
    ComboBoxColumnDescription.prototype.get_type = function () {
        return "ComboBoxColumn";
    };
    Object.defineProperty(ComboBoxColumnDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboBoxColumnDescription.prototype, "dataSource", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("DataSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboBoxColumnDescription.prototype, "textField", {
        get: function () {
            return this.f6;
        },
        set: function (a) {
            this.f6 = a;
            this.g("TextField");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ComboBoxColumnDescription.prototype, "valueField", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("ValueField");
        },
        enumerable: false,
        configurable: true
    });
    ComboBoxColumnDescription.$t = markType(ComboBoxColumnDescription, 'ComboBoxColumnDescription', DataGridColumnDescription.$);
    ComboBoxColumnDescription.__canMarshalByValue1 = true;
    ComboBoxColumnDescription.__marshalByValueAlias1 = "ComboBoxColumn";
    return ComboBoxColumnDescription;
}(DataGridColumnDescription));
export { ComboBoxColumnDescription };
