// igniteui-react-core/esm5/lib/DashboardTileCustomizationDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DashboardTileCustomizationDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DashboardTileCustomizationDescription, _super);
    function DashboardTileCustomizationDescription() {
        var _this = _super.call(this) || this;
        _this.v = null;
        _this.w = null;
        _this.x = null;
        _this.h = false;
        _this.j = false;
        _this.i = false;
        _this.k = false;
        _this.p = 0;
        _this.r = 0;
        _this.q = 0;
        return _this;
    }
    DashboardTileCustomizationDescription.prototype.get_type = function () {
        return "DashboardTileCustomization";
    };
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "matchType", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("MatchType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "propertyName", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("PropertyName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "propertyValueRef", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("PropertyValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "isCollectionInsertionAtEnd", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("IsCollectionInsertionAtEnd");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "isCollectionInsertionAtStart", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("IsCollectionInsertionAtStart");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "isCollectionInsertionAtIndex", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("IsCollectionInsertionAtIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "isCollectionRemovaltIndex", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("IsCollectionRemovaltIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "collectionIndex", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("CollectionIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "order", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Order");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashboardTileCustomizationDescription.prototype, "matchParentIndex", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("MatchParentIndex");
        },
        enumerable: false,
        configurable: true
    });
    DashboardTileCustomizationDescription.$t = markType(DashboardTileCustomizationDescription, 'DashboardTileCustomizationDescription', Description.$);
    return DashboardTileCustomizationDescription;
}(Description));
export { DashboardTileCustomizationDescription };
