// igniteui-react-core/esm5/lib/DataContextDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataContextDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataContextDescription, _super);
    function DataContextDescription() {
        var _this = _super.call(this) || this;
        _this.o = null;
        _this.l = null;
        _this.q = null;
        _this.n = null;
        _this.m = null;
        _this.i = 0;
        _this.p = null;
        return _this;
    }
    DataContextDescription.prototype.get_type = function () {
        return "DataContext";
    };
    Object.defineProperty(DataContextDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContextDescription.prototype, "itemRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContextDescription.prototype, "actualItemBrush", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ActualItemBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContextDescription.prototype, "outline", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Outline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContextDescription.prototype, "itemLabelRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ItemLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContextDescription.prototype, "itemBrush", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("ItemBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContextDescription.prototype, "thickness", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Thickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContextDescription.prototype, "legendLabelRef", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("LegendLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    DataContextDescription.$t = markType(DataContextDescription, 'DataContextDescription', Description.$);
    DataContextDescription.__marshalByValue = true;
    DataContextDescription.__marshalByValueAlias = "DataContext";
    return DataContextDescription;
}(Description));
export { DataContextDescription };
