// igniteui-react-core/esm5/lib/DataVisualizationLocaleHu.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocaleHu = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocaleHu() {
        this["AssemblyNotIncluded"] = "A(z) {0}.dll nem tölthető be. Győződjön meg arról, hogy ez az összeállítás szerepel-e a Windows Phone-ra telepített XAP-ben, például úgy, hogy hozzáad egy arra mutató hivatkozást bármelyik projektjéhez.";
        this["BreakEvenTitle_BreakEven"] = "Nullszaldó";
        this["BreakEvenTitle_FixedCost"] = "Fix költség";
        this["BreakEvenTitle_LossArea"] = "Veszteségterület";
        this["BreakEvenTitle_MarginalProfit"] = "Határnyereség";
        this["BreakEvenTitle_ProfitArea"] = "Nyereségterület";
        this["BreakEvenTitle_SafetyMargin"] = "Biztonsági margin";
        this["BreakEvenTitle_SalesRevenue"] = "Árbevétel";
        this["BreakEvenTitle_TotalCost"] = "Összköltség";
        this["BreakEvenTitle_VariableCost"] = "Változó költség";
        this["BubbleSeries_Radius"] = "Sugár";
        this["DataChart_InteractivityNotLoaded"] = "Az interaktivitási funkció meghívásra került, de a diagram-interakciós modul nem töltődött be.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "A számrövidítési funkció meghívásra került, de a számrövidítő modul nem töltődött be.";
        this["DataChart_VisualDataNotLoaded"] = "A VisualData funkció meghívásra került, de a diagram-interakciós modul nem töltődött be.";
        this["DataProviderNotInitialized"] = "Az adatszolgáltatót inicializálás nélkül használták. Az adatszolgáltatót inicializálni kell használat előtt.";
        this["DataSource_Summary_Avg"] = "Átlag";
        this["DataSource_Summary_Count"] = "Darab";
        this["DataSource_Summary_Max"] = "Maximum";
        this["DataSource_Summary_Min"] = "Minimum";
        this["DataSource_Summary_Sum"] = "Szum";
        this["Default_Series_Title"] = "Adatsor címe";
        this["FinancialChart_IndicatorMenu_Header"] = "Mutatók";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Önálló mutatók";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Rétegek";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Trendvonalak";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volumenek";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Abszolút volumen oszcillátor";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Kumulatív/Eloszlás";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Átlagolt irányindex";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Átlagolt valós tartomány";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Bollinger szalag szélessége";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Chaikin oszcillátor";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Chaikin volatilitás";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Árucsatorna-index";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Trendmentesített ároszcillátor";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Árfolyam nyugodtsága";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Gyors sztochasztikus oszcillátor";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Force Index";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Teljes sztochasztikus oszcillátor";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Piackönnyítési index";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Volumen index";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Medián ár";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Pénzforgalmi index";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Negatív volumen index";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "Egyensúlyi volumen";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Százalékos ár oszcillátor";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Százalékos volumen oszcillátor";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Pozitív volumen index";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Ár-volumen trend";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Lassú sztochasztikus oszcillátor";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Szórás";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Sztochasztikus RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Tipikus ár";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Végső oszcillátor";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Súlyozott záróérték";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bollinger szalagok";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Árcsatorna";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Harmadfokú illesztés";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Halmozott átlag";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Exponenciális mozgóátlag";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Exponenciális illesztés";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Lineáris illesztés";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Logaritmikus illesztés";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Módosított átlag";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "Hatványfüggvényes illesztés";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Másodfokú illesztés";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Negyedfokú illesztés";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Ötödfokú illesztés";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Számtani átlag";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Súlyozott átlag";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Terület";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Oszlop";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Vonal";
        this["FinancialChart_RangeSelector_All"] = "Összes";
        this["FinancialChart_RangeSelector_From"] = "Ettől";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 hó";
        this["FinancialChart_RangeSelector_OneYear"] = "1 év";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 hó";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 hó";
        this["FinancialChart_RangeSelector_To"] = "Eddig";
        this["FinancialChart_RangeSelector_YearToDate"] = "Tárgyév";
        this["FinancialSeries_Close"] = "Záró";
        this["FinancialSeries_High"] = "Legmagasabb";
        this["FinancialSeries_Low"] = "Legalacsonyabb";
        this["FinancialSeries_Open"] = "Nyitó";
        this["FinancialSeries_Volume"] = "Volumen";
        this["NoEncodingsLoaded"] = "Nem találhatók kódolások! Töltse be az Infragistics.Encoding összeállítást, vagy nézzen utána a dokumentációban, hogy miként lehet létrehozni egyéni kódolást.";
        this["NotSupportedEncoding"] = "A(z) {0} {1} alatt található kódolás nem támogatott.";
        this["Object_Sealed"] = "Az objektum lezárása után nem végezhetők módosítások.";
        this["OPD_DefaultInteraction"] = "Válassza ki az egérhúzáshoz rendelt alapértelmezett műveletet";
        this["OPD_DragPan"] = "Nézet csúsztatása";
        this["OPD_DragZoom"] = "Nagyítás a kijelölésre";
        this["OPD_ScaleToFit"] = "Képernyőhöz igazítás";
        this["OPD_ScaleToFit_SeriesViewer"] = "Nagyítás visszaállítása (100%)";
        this["OPD_ZoomIn"] = "Nagyítás";
        this["OPD_ZoomOut"] = "Kicsinyítés";
        this["OPD_ZoomTo100"] = "Nagyítás 100%-ra";
        this["PieChart_Others"] = "Egyebek";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Az adatforrás tartományalapú módosítása jelenleg nem támogatott";
        this["ScatterSeries_Value"] = "Érték";
        this["TRIAL_VERSION"] = "LICENC NÉLKÜLI PRÓBAVERZIÓ Vásárlás az Infragistics.com oldalon";
    }
    return DataVisualizationLocaleHu;
}());
export { DataVisualizationLocaleHu };
