// igniteui-react-core/esm5/lib/DataToolTipLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var DataToolTipLayerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataToolTipLayerDescription, _super);
    function DataToolTipLayerDescription() {
        var _this = _super.call(this) || this;
        _this.lv = null;
        _this.go = false;
        _this.kr = null;
        _this.kn = null;
        _this.hs = 0;
        _this.ht = 0;
        _this.g3 = 0;
        _this.g4 = 0;
        _this.kp = null;
        _this.kl = null;
        _this.kq = null;
        _this.km = null;
        _this.gn = false;
        _this.gd = null;
        _this.ga = null;
        _this.gc = null;
        _this.f9 = null;
        _this.l6 = null;
        _this.kh = 0;
        _this.ki = 0;
        _this.l8 = null;
        _this.l7 = null;
        _this.gp = false;
        _this.l9 = null;
        _this.ge = null;
        _this.ih = 0;
        _this.ii = 0;
        _this.ij = 0;
        _this.ik = 0;
        _this.gq = false;
        _this.me = null;
        _this.gr = false;
        _this.im = 0;
        _this.io = 0;
        _this.ip = 0;
        _this.iq = 0;
        _this.ma = null;
        _this.mb = null;
        _this.il = 0;
        _this.mc = null;
        _this.md = null;
        _this.ky = null;
        _this.gb = null;
        _this.kw = null;
        _this.kx = null;
        _this.kz = null;
        _this.k0 = null;
        _this.k1 = null;
        _this.hj = 0;
        _this.hk = 0;
        _this.hl = 0;
        _this.hm = 0;
        _this.he = 0;
        _this.hf = 0;
        _this.hg = 0;
        _this.hh = 0;
        _this.gm = false;
        _this.k2 = null;
        _this.hi = 0;
        _this.k3 = null;
        _this.k4 = null;
        _this.gl = false;
        _this.ks = null;
        _this.ha = 0;
        _this.hb = 0;
        _this.hc = 0;
        _this.hd = 0;
        _this.g5 = 0;
        _this.g6 = 0;
        _this.g7 = 0;
        _this.g8 = 0;
        _this.kt = null;
        _this.g9 = 0;
        _this.ku = null;
        _this.kv = null;
        _this.lh = null;
        _this.li = null;
        _this.hz = 0;
        _this.lj = null;
        _this.lk = null;
        _this.ll = null;
        _this.lg = null;
        _this.h0 = 0;
        _this.h1 = 0;
        _this.h2 = 0;
        _this.h3 = 0;
        _this.hv = 0;
        _this.hw = 0;
        _this.hx = 0;
        _this.hy = 0;
        _this.lr = null;
        _this.ls = null;
        _this.h5 = 0;
        _this.lt = null;
        _this.lu = null;
        _this.lb = null;
        _this.lc = null;
        _this.ld = null;
        _this.hu = 0;
        _this.le = null;
        _this.lf = null;
        _this.lm = null;
        _this.ln = null;
        _this.lo = null;
        _this.h4 = 0;
        _this.lp = null;
        _this.lq = null;
        _this.gz = 0;
        _this.g0 = 0;
        _this.g1 = 0;
        _this.g2 = 0;
        _this.ko = null;
        _this.la = null;
        _this.l0 = null;
        _this.l1 = null;
        _this.l2 = null;
        _this.ic = 0;
        _this.id = 0;
        _this.ie = 0;
        _this.ig = 0;
        _this.l3 = null;
        _this.ib = 0;
        _this.l4 = null;
        _this.l5 = null;
        _this.h7 = 0;
        _this.h8 = 0;
        _this.h9 = 0;
        _this.ia = 0;
        _this.lw = null;
        _this.lx = null;
        _this.h6 = 0;
        _this.ly = null;
        _this.lz = null;
        _this.k5 = null;
        _this.k6 = null;
        _this.ho = 0;
        _this.hp = 0;
        _this.hq = 0;
        _this.hr = 0;
        _this.k7 = null;
        _this.hn = 0;
        _this.k8 = null;
        _this.k9 = null;
        return _this;
    }
    DataToolTipLayerDescription.prototype.get_type = function () {
        return "DataToolTipLayer";
    };
    Object.defineProperty(DataToolTipLayerDescription.prototype, "targetAxisRef", {
        get: function () {
            return this.lv;
        },
        set: function (a) {
            this.lv = a;
            this.g("TargetAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "useInterpolation", {
        get: function () {
            return this.go;
        },
        set: function (a) {
            this.go = a;
            this.g("UseInterpolation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupingMode", {
        get: function () {
            return this.kr;
        },
        set: function (a) {
            this.kr = a;
            this.g("GroupingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "actualGroupingMode", {
        get: function () {
            return this.kn;
        },
        set: function (a) {
            this.kn = a;
            this.g("ActualGroupingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "positionOffsetX", {
        get: function () {
            return this.hs;
        },
        set: function (a) {
            this.hs = a;
            this.g("PositionOffsetX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "positionOffsetY", {
        get: function () {
            return this.ht;
        },
        set: function (a) {
            this.ht = a;
            this.g("PositionOffsetY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "defaultPositionOffsetX", {
        get: function () {
            return this.g3;
        },
        set: function (a) {
            this.g3 = a;
            this.g("DefaultPositionOffsetX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "defaultPositionOffsetY", {
        get: function () {
            return this.g4;
        },
        set: function (a) {
            this.g4 = a;
            this.g("DefaultPositionOffsetY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupedPositionModeX", {
        get: function () {
            return this.kp;
        },
        set: function (a) {
            this.kp = a;
            this.g("GroupedPositionModeX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "actualGroupedPositionModeX", {
        get: function () {
            return this.kl;
        },
        set: function (a) {
            this.kl = a;
            this.g("ActualGroupedPositionModeX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupedPositionModeY", {
        get: function () {
            return this.kq;
        },
        set: function (a) {
            this.kq = a;
            this.g("GroupedPositionModeY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "actualGroupedPositionModeY", {
        get: function () {
            return this.km;
        },
        set: function (a) {
            this.km = a;
            this.g("ActualGroupedPositionModeY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "shouldUpdateWhenSeriesDataChanges", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("ShouldUpdateWhenSeriesDataChanges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "includedSeries", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("IncludedSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "excludedSeries", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("ExcludedSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "includedColumns", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("IncludedColumns");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "excludedColumns", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("ExcludedColumns");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatAbbreviation", {
        get: function () {
            return this.l6;
        },
        set: function (a) {
            this.l6 = a;
            this.g("ValueFormatAbbreviation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatMaxFractions", {
        get: function () {
            return this.kh;
        },
        set: function (a) {
            this.kh = a;
            this.g("ValueFormatMaxFractions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatMinFractions", {
        get: function () {
            return this.ki;
        },
        set: function (a) {
            this.ki = a;
            this.g("ValueFormatMinFractions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatMode", {
        get: function () {
            return this.l8;
        },
        set: function (a) {
            this.l8 = a;
            this.g("ValueFormatMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatCulture", {
        get: function () {
            return this.l7;
        },
        set: function (a) {
            this.l7 = a;
            this.g("ValueFormatCulture");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatUseGrouping", {
        get: function () {
            return this.gp;
        },
        set: function (a) {
            this.gp = a;
            this.g("ValueFormatUseGrouping");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatString", {
        get: function () {
            return this.l9;
        },
        set: function (a) {
            this.l9 = a;
            this.g("ValueFormatString");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueFormatSpecifiers", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("ValueFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueRowMarginBottom", {
        get: function () {
            return this.ih;
        },
        set: function (a) {
            this.ih = a;
            this.g("ValueRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueRowMarginLeft", {
        get: function () {
            return this.ii;
        },
        set: function (a) {
            this.ii = a;
            this.g("ValueRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueRowMarginRight", {
        get: function () {
            return this.ij;
        },
        set: function (a) {
            this.ij = a;
            this.g("ValueRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueRowMarginTop", {
        get: function () {
            return this.ik;
        },
        set: function (a) {
            this.ik = a;
            this.g("ValueRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueRowVisible", {
        get: function () {
            return this.gq;
        },
        set: function (a) {
            this.gq = a;
            this.g("ValueRowVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextWhenMissingData", {
        get: function () {
            return this.me;
        },
        set: function (a) {
            this.me = a;
            this.g("ValueTextWhenMissingData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextUseSeriesColors", {
        get: function () {
            return this.gr;
        },
        set: function (a) {
            this.gr = a;
            this.g("ValueTextUseSeriesColors");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextMarginBottom", {
        get: function () {
            return this.im;
        },
        set: function (a) {
            this.im = a;
            this.g("ValueTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextMarginLeft", {
        get: function () {
            return this.io;
        },
        set: function (a) {
            this.io = a;
            this.g("ValueTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextMarginRight", {
        get: function () {
            return this.ip;
        },
        set: function (a) {
            this.ip = a;
            this.g("ValueTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextMarginTop", {
        get: function () {
            return this.iq;
        },
        set: function (a) {
            this.iq = a;
            this.g("ValueTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextColor", {
        get: function () {
            return this.ma;
        },
        set: function (a) {
            this.ma = a;
            this.g("ValueTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextFontFamily", {
        get: function () {
            return this.mb;
        },
        set: function (a) {
            this.mb = a;
            this.g("ValueTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextFontSize", {
        get: function () {
            return this.il;
        },
        set: function (a) {
            this.il = a;
            this.g("ValueTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextFontStyle", {
        get: function () {
            return this.mc;
        },
        set: function (a) {
            this.mc = a;
            this.g("ValueTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "valueTextFontWeight", {
        get: function () {
            return this.md;
        },
        set: function (a) {
            this.md = a;
            this.g("ValueTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerFormatString", {
        get: function () {
            return this.ky;
        },
        set: function (a) {
            this.ky = a;
            this.g("HeaderFormatString");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerFormatSpecifiers", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("HeaderFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerFormatCulture", {
        get: function () {
            return this.kw;
        },
        set: function (a) {
            this.kw = a;
            this.g("HeaderFormatCulture");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerFormatDate", {
        get: function () {
            return this.kx;
        },
        set: function (a) {
            this.kx = a;
            this.g("HeaderFormatDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerFormatTime", {
        get: function () {
            return this.kz;
        },
        set: function (a) {
            this.kz = a;
            this.g("HeaderFormatTime");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerText", {
        get: function () {
            return this.k0;
        },
        set: function (a) {
            this.k0 = a;
            this.g("HeaderText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextColor", {
        get: function () {
            return this.k1;
        },
        set: function (a) {
            this.k1 = a;
            this.g("HeaderTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextMarginBottom", {
        get: function () {
            return this.hj;
        },
        set: function (a) {
            this.hj = a;
            this.g("HeaderTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextMarginLeft", {
        get: function () {
            return this.hk;
        },
        set: function (a) {
            this.hk = a;
            this.g("HeaderTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextMarginRight", {
        get: function () {
            return this.hl;
        },
        set: function (a) {
            this.hl = a;
            this.g("HeaderTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextMarginTop", {
        get: function () {
            return this.hm;
        },
        set: function (a) {
            this.hm = a;
            this.g("HeaderTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerRowMarginBottom", {
        get: function () {
            return this.he;
        },
        set: function (a) {
            this.he = a;
            this.g("HeaderRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerRowMarginLeft", {
        get: function () {
            return this.hf;
        },
        set: function (a) {
            this.hf = a;
            this.g("HeaderRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerRowMarginRight", {
        get: function () {
            return this.hg;
        },
        set: function (a) {
            this.hg = a;
            this.g("HeaderRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerRowMarginTop", {
        get: function () {
            return this.hh;
        },
        set: function (a) {
            this.hh = a;
            this.g("HeaderRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerRowVisible", {
        get: function () {
            return this.gm;
        },
        set: function (a) {
            this.gm = a;
            this.g("HeaderRowVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextFontFamily", {
        get: function () {
            return this.k2;
        },
        set: function (a) {
            this.k2 = a;
            this.g("HeaderTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextFontSize", {
        get: function () {
            return this.hi;
        },
        set: function (a) {
            this.hi = a;
            this.g("HeaderTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextFontStyle", {
        get: function () {
            return this.k3;
        },
        set: function (a) {
            this.k3 = a;
            this.g("HeaderTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "headerTextFontWeight", {
        get: function () {
            return this.k4;
        },
        set: function (a) {
            this.k4 = a;
            this.g("HeaderTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupRowVisible", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("GroupRowVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupTextColor", {
        get: function () {
            return this.ks;
        },
        set: function (a) {
            this.ks = a;
            this.g("GroupTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupTextMarginBottom", {
        get: function () {
            return this.ha;
        },
        set: function (a) {
            this.ha = a;
            this.g("GroupTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupTextMarginLeft", {
        get: function () {
            return this.hb;
        },
        set: function (a) {
            this.hb = a;
            this.g("GroupTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupTextMarginRight", {
        get: function () {
            return this.hc;
        },
        set: function (a) {
            this.hc = a;
            this.g("GroupTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupTextMarginTop", {
        get: function () {
            return this.hd;
        },
        set: function (a) {
            this.hd = a;
            this.g("GroupTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupRowMarginBottom", {
        get: function () {
            return this.g5;
        },
        set: function (a) {
            this.g5 = a;
            this.g("GroupRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupRowMarginLeft", {
        get: function () {
            return this.g6;
        },
        set: function (a) {
            this.g6 = a;
            this.g("GroupRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupRowMarginRight", {
        get: function () {
            return this.g7;
        },
        set: function (a) {
            this.g7 = a;
            this.g("GroupRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupRowMarginTop", {
        get: function () {
            return this.g8;
        },
        set: function (a) {
            this.g8 = a;
            this.g("GroupRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupTextFontFamily", {
        get: function () {
            return this.kt;
        },
        set: function (a) {
            this.kt = a;
            this.g("GroupTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupTextFontSize", {
        get: function () {
            return this.g9;
        },
        set: function (a) {
            this.g9 = a;
            this.g("GroupTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupTextFontStyle", {
        get: function () {
            return this.ku;
        },
        set: function (a) {
            this.ku = a;
            this.g("GroupTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "groupTextFontWeight", {
        get: function () {
            return this.kv;
        },
        set: function (a) {
            this.kv = a;
            this.g("GroupTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextColor", {
        get: function () {
            return this.lh;
        },
        set: function (a) {
            this.lh = a;
            this.g("SummaryTitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextFontFamily", {
        get: function () {
            return this.li;
        },
        set: function (a) {
            this.li = a;
            this.g("SummaryTitleTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextFontSize", {
        get: function () {
            return this.hz;
        },
        set: function (a) {
            this.hz = a;
            this.g("SummaryTitleTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextFontStyle", {
        get: function () {
            return this.lj;
        },
        set: function (a) {
            this.lj = a;
            this.g("SummaryTitleTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextFontWeight", {
        get: function () {
            return this.lk;
        },
        set: function (a) {
            this.lk = a;
            this.g("SummaryTitleTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryType", {
        get: function () {
            return this.ll;
        },
        set: function (a) {
            this.ll = a;
            this.g("SummaryType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleText", {
        get: function () {
            return this.lg;
        },
        set: function (a) {
            this.lg = a;
            this.g("SummaryTitleText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextMarginBottom", {
        get: function () {
            return this.h0;
        },
        set: function (a) {
            this.h0 = a;
            this.g("SummaryTitleTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextMarginLeft", {
        get: function () {
            return this.h1;
        },
        set: function (a) {
            this.h1 = a;
            this.g("SummaryTitleTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextMarginRight", {
        get: function () {
            return this.h2;
        },
        set: function (a) {
            this.h2 = a;
            this.g("SummaryTitleTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryTitleTextMarginTop", {
        get: function () {
            return this.h3;
        },
        set: function (a) {
            this.h3 = a;
            this.g("SummaryTitleTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryRowMarginBottom", {
        get: function () {
            return this.hv;
        },
        set: function (a) {
            this.hv = a;
            this.g("SummaryRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryRowMarginLeft", {
        get: function () {
            return this.hw;
        },
        set: function (a) {
            this.hw = a;
            this.g("SummaryRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryRowMarginRight", {
        get: function () {
            return this.hx;
        },
        set: function (a) {
            this.hx = a;
            this.g("SummaryRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryRowMarginTop", {
        get: function () {
            return this.hy;
        },
        set: function (a) {
            this.hy = a;
            this.g("SummaryRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryValueTextColor", {
        get: function () {
            return this.lr;
        },
        set: function (a) {
            this.lr = a;
            this.g("SummaryValueTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryValueTextFontFamily", {
        get: function () {
            return this.ls;
        },
        set: function (a) {
            this.ls = a;
            this.g("SummaryValueTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryValueTextFontSize", {
        get: function () {
            return this.h5;
        },
        set: function (a) {
            this.h5 = a;
            this.g("SummaryValueTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryValueTextFontStyle", {
        get: function () {
            return this.lt;
        },
        set: function (a) {
            this.lt = a;
            this.g("SummaryValueTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryValueTextFontWeight", {
        get: function () {
            return this.lu;
        },
        set: function (a) {
            this.lu = a;
            this.g("SummaryValueTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelText", {
        get: function () {
            return this.lb;
        },
        set: function (a) {
            this.lb = a;
            this.g("SummaryLabelText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelTextColor", {
        get: function () {
            return this.lc;
        },
        set: function (a) {
            this.lc = a;
            this.g("SummaryLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelTextFontFamily", {
        get: function () {
            return this.ld;
        },
        set: function (a) {
            this.ld = a;
            this.g("SummaryLabelTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelTextFontSize", {
        get: function () {
            return this.hu;
        },
        set: function (a) {
            this.hu = a;
            this.g("SummaryLabelTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelTextFontStyle", {
        get: function () {
            return this.le;
        },
        set: function (a) {
            this.le = a;
            this.g("SummaryLabelTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryLabelTextFontWeight", {
        get: function () {
            return this.lf;
        },
        set: function (a) {
            this.lf = a;
            this.g("SummaryLabelTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsText", {
        get: function () {
            return this.lm;
        },
        set: function (a) {
            this.lm = a;
            this.g("SummaryUnitsText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsTextColor", {
        get: function () {
            return this.ln;
        },
        set: function (a) {
            this.ln = a;
            this.g("SummaryUnitsTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsTextFontFamily", {
        get: function () {
            return this.lo;
        },
        set: function (a) {
            this.lo = a;
            this.g("SummaryUnitsTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsTextFontSize", {
        get: function () {
            return this.h4;
        },
        set: function (a) {
            this.h4 = a;
            this.g("SummaryUnitsTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsTextFontStyle", {
        get: function () {
            return this.lp;
        },
        set: function (a) {
            this.lp = a;
            this.g("SummaryUnitsTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "summaryUnitsTextFontWeight", {
        get: function () {
            return this.lq;
        },
        set: function (a) {
            this.lq = a;
            this.g("SummaryUnitsTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "badgeMarginBottom", {
        get: function () {
            return this.gz;
        },
        set: function (a) {
            this.gz = a;
            this.g("BadgeMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "badgeMarginLeft", {
        get: function () {
            return this.g0;
        },
        set: function (a) {
            this.g0 = a;
            this.g("BadgeMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "badgeMarginRight", {
        get: function () {
            return this.g1;
        },
        set: function (a) {
            this.g1 = a;
            this.g("BadgeMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "badgeMarginTop", {
        get: function () {
            return this.g2;
        },
        set: function (a) {
            this.g2 = a;
            this.g("BadgeMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "badgeShape", {
        get: function () {
            return this.ko;
        },
        set: function (a) {
            this.ko = a;
            this.g("BadgeShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "layoutMode", {
        get: function () {
            return this.la;
        },
        set: function (a) {
            this.la = a;
            this.g("LayoutMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsDisplayMode", {
        get: function () {
            return this.l0;
        },
        set: function (a) {
            this.l0 = a;
            this.g("UnitsDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsText", {
        get: function () {
            return this.l1;
        },
        set: function (a) {
            this.l1 = a;
            this.g("UnitsText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextColor", {
        get: function () {
            return this.l2;
        },
        set: function (a) {
            this.l2 = a;
            this.g("UnitsTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextMarginBottom", {
        get: function () {
            return this.ic;
        },
        set: function (a) {
            this.ic = a;
            this.g("UnitsTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextMarginLeft", {
        get: function () {
            return this.id;
        },
        set: function (a) {
            this.id = a;
            this.g("UnitsTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextMarginRight", {
        get: function () {
            return this.ie;
        },
        set: function (a) {
            this.ie = a;
            this.g("UnitsTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextMarginTop", {
        get: function () {
            return this.ig;
        },
        set: function (a) {
            this.ig = a;
            this.g("UnitsTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextFontFamily", {
        get: function () {
            return this.l3;
        },
        set: function (a) {
            this.l3 = a;
            this.g("UnitsTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextFontSize", {
        get: function () {
            return this.ib;
        },
        set: function (a) {
            this.ib = a;
            this.g("UnitsTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextFontStyle", {
        get: function () {
            return this.l4;
        },
        set: function (a) {
            this.l4 = a;
            this.g("UnitsTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "unitsTextFontWeight", {
        get: function () {
            return this.l5;
        },
        set: function (a) {
            this.l5 = a;
            this.g("UnitsTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextMarginBottom", {
        get: function () {
            return this.h7;
        },
        set: function (a) {
            this.h7 = a;
            this.g("TitleTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextMarginLeft", {
        get: function () {
            return this.h8;
        },
        set: function (a) {
            this.h8 = a;
            this.g("TitleTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextMarginRight", {
        get: function () {
            return this.h9;
        },
        set: function (a) {
            this.h9 = a;
            this.g("TitleTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextMarginTop", {
        get: function () {
            return this.ia;
        },
        set: function (a) {
            this.ia = a;
            this.g("TitleTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextColor", {
        get: function () {
            return this.lw;
        },
        set: function (a) {
            this.lw = a;
            this.g("TitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextFontFamily", {
        get: function () {
            return this.lx;
        },
        set: function (a) {
            this.lx = a;
            this.g("TitleTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextFontSize", {
        get: function () {
            return this.h6;
        },
        set: function (a) {
            this.h6 = a;
            this.g("TitleTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextFontStyle", {
        get: function () {
            return this.ly;
        },
        set: function (a) {
            this.ly = a;
            this.g("TitleTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "titleTextFontWeight", {
        get: function () {
            return this.lz;
        },
        set: function (a) {
            this.lz = a;
            this.g("TitleTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelDisplayMode", {
        get: function () {
            return this.k5;
        },
        set: function (a) {
            this.k5 = a;
            this.g("LabelDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextColor", {
        get: function () {
            return this.k6;
        },
        set: function (a) {
            this.k6 = a;
            this.g("LabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextMarginBottom", {
        get: function () {
            return this.ho;
        },
        set: function (a) {
            this.ho = a;
            this.g("LabelTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextMarginLeft", {
        get: function () {
            return this.hp;
        },
        set: function (a) {
            this.hp = a;
            this.g("LabelTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextMarginRight", {
        get: function () {
            return this.hq;
        },
        set: function (a) {
            this.hq = a;
            this.g("LabelTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextMarginTop", {
        get: function () {
            return this.hr;
        },
        set: function (a) {
            this.hr = a;
            this.g("LabelTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextFontFamily", {
        get: function () {
            return this.k7;
        },
        set: function (a) {
            this.k7 = a;
            this.g("LabelTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextFontSize", {
        get: function () {
            return this.hn;
        },
        set: function (a) {
            this.hn = a;
            this.g("LabelTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextFontStyle", {
        get: function () {
            return this.k8;
        },
        set: function (a) {
            this.k8 = a;
            this.g("LabelTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayerDescription.prototype, "labelTextFontWeight", {
        get: function () {
            return this.k9;
        },
        set: function (a) {
            this.k9 = a;
            this.g("LabelTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    DataToolTipLayerDescription.$t = markType(DataToolTipLayerDescription, 'DataToolTipLayerDescription', AnnotationLayerDescription.$);
    return DataToolTipLayerDescription;
}(AnnotationLayerDescription));
export { DataToolTipLayerDescription };
