// igniteui-react-core/esm5/lib/DataGridPagerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataGridPagerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataGridPagerDescription, _super);
    function DataGridPagerDescription() {
        var _this = _super.call(this) || this;
        _this.y = null;
        _this.n = 0;
        _this.m = 0;
        _this.t = null;
        _this.h = null;
        _this.l = 0;
        _this.u = null;
        _this.j = 0;
        _this.v = null;
        _this.w = null;
        _this.r = null;
        _this.s = null;
        _this.x = null;
        return _this;
    }
    DataGridPagerDescription.prototype.get_type = function () {
        return "DataGridPager";
    };
    Object.defineProperty(DataGridPagerDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "pagerText", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("PagerText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "pageSize", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("PageSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "pageCount", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("PageCount");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "dataSource", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("DataSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "currentPage", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("CurrentPage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "fontFamily", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("FontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "fontSize", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("FontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "fontStyle", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("FontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "fontWeight", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("FontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "background", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "border", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Border");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "pageChangedRef", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("PageChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    DataGridPagerDescription.$t = markType(DataGridPagerDescription, 'DataGridPagerDescription', Description.$);
    return DataGridPagerDescription;
}(Description));
export { DataGridPagerDescription };
