// igniteui-react-core/esm5/lib/DataLegendStylingColumnEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataLegendStylingColumnEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataLegendStylingColumnEventArgsDescription, _super);
    function DataLegendStylingColumnEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.j = 0;
        _this.k = 0;
        _this.q = null;
        _this.h = 0;
        _this.v = null;
        _this.u = null;
        _this.t = null;
        _this.o = null;
        _this.r = null;
        _this.w = null;
        _this.p = null;
        _this.s = null;
        _this.x = null;
        _this.n = null;
        return _this;
    }
    DataLegendStylingColumnEventArgsDescription.prototype.get_type = function () {
        return "DataLegendStylingColumnEventArgs";
    };
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "columnIndex", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ColumnIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "seriesIndex", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("SeriesIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "seriesTitle", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("SeriesTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "valueOriginal", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ValueOriginal");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "valueMemberLabel", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("ValueMemberLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "valueAbbreviation", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("ValueAbbreviation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "labelText", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("LabelText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "unitsText", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("UnitsText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "valueText", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("ValueText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "labelTextColor", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("LabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "unitsTextColor", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("UnitsTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "valueTextColor", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("ValueTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingColumnEventArgsDescription.prototype, "groupName", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("GroupName");
        },
        enumerable: false,
        configurable: true
    });
    DataLegendStylingColumnEventArgsDescription.$t = markType(DataLegendStylingColumnEventArgsDescription, 'DataLegendStylingColumnEventArgsDescription', Description.$);
    return DataLegendStylingColumnEventArgsDescription;
}(Description));
export { DataLegendStylingColumnEventArgsDescription };
