// igniteui-react-core/esm5/lib/DataLegendStylingRowEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataLegendStylingRowEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataLegendStylingRowEventArgsDescription, _super);
    function DataLegendStylingRowEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.h = false;
        _this.i = false;
        _this.l = 0;
        _this.p = null;
        _this.q = null;
        _this.o = null;
        _this.r = null;
        return _this;
    }
    DataLegendStylingRowEventArgsDescription.prototype.get_type = function () {
        return "DataLegendStylingRowEventArgs";
    };
    Object.defineProperty(DataLegendStylingRowEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingRowEventArgsDescription.prototype, "badgeShape", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("BadgeShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingRowEventArgsDescription.prototype, "isBadgeVisible", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("IsBadgeVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingRowEventArgsDescription.prototype, "isRowVisible", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("IsRowVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingRowEventArgsDescription.prototype, "seriesIndex", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("SeriesIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingRowEventArgsDescription.prototype, "seriesTitle", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("SeriesTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingRowEventArgsDescription.prototype, "titleText", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("TitleText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingRowEventArgsDescription.prototype, "groupName", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("GroupName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendStylingRowEventArgsDescription.prototype, "titleTextColor", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("TitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    DataLegendStylingRowEventArgsDescription.$t = markType(DataLegendStylingRowEventArgsDescription, 'DataLegendStylingRowEventArgsDescription', Description.$);
    return DataLegendStylingRowEventArgsDescription;
}(Description));
export { DataLegendStylingRowEventArgsDescription };
