// igniteui-react-core/esm5/lib/DataGridToolbarDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataGridToolbarDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataGridToolbarDescription, _super);
    function DataGridToolbarDescription() {
        var _this = _super.call(this) || this;
        _this.ad = null;
        _this.ae = null;
        _this.ag = null;
        _this.p = 0;
        _this.ah = null;
        _this.ai = null;
        _this.af = null;
        _this.v = null;
        _this.ac = null;
        _this.h = false;
        _this.x = null;
        _this.y = null;
        _this.i = false;
        _this.z = null;
        _this.aa = null;
        _this.l = 0;
        _this.m = 0;
        _this.n = 0;
        _this.o = 0;
        _this.ab = null;
        _this.w = null;
        return _this;
    }
    DataGridToolbarDescription.prototype.get_type = function () {
        return "DataGridToolbar";
    };
    Object.defineProperty(DataGridToolbarDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "targetGridRef", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("TargetGridRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "toolbarTitle", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("ToolbarTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "toolbarTitleFontFamily", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("ToolbarTitleFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "toolbarTitleFontSize", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("ToolbarTitleFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "toolbarTitleFontStyle", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("ToolbarTitleFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "toolbarTitleFontWeight", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("ToolbarTitleFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "toolbarTitleColor", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("ToolbarTitleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "backgroundColor", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("BackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "dialogBackgroundColor", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("DialogBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "columnChooser", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ColumnChooser");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "columnChooserText", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("ColumnChooserText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "columnChooserTitle", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("ColumnChooserTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "columnPinning", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ColumnPinning");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "columnPinningText", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("ColumnPinningText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "columnPinningTitle", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("ColumnPinningTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "borderWidthBottom", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("BorderWidthBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "borderWidthLeft", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("BorderWidthLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "borderWidthRight", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("BorderWidthRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "borderWidthTop", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("BorderWidthTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "density", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("Density");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridToolbarDescription.prototype, "baseTheme", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    DataGridToolbarDescription.$t = markType(DataGridToolbarDescription, 'DataGridToolbarDescription', Description.$);
    return DataGridToolbarDescription;
}(Description));
export { DataGridToolbarDescription };
