// igniteui-react-core/esm5/lib/DataGridSortIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataGridSortIndicatorDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataGridSortIndicatorDescription, _super);
    function DataGridSortIndicatorDescription() {
        var _this = _super.call(this) || this;
        _this.p = null;
        _this.n = null;
        _this.h = false;
        _this.k = 0;
        _this.j = 0;
        _this.q = null;
        _this.o = null;
        return _this;
    }
    DataGridSortIndicatorDescription.prototype.get_type = function () {
        return "DataGridSortIndicator";
    };
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "sortDirection", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("SortDirection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "iconColor", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("IconColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "isAnimationEnabled", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("IsAnimationEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("PixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "actualPixelScalingRatio", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ActualPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "sortIndicatorStyle", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("SortIndicatorStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "renderCompletedRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("RenderCompletedRef");
        },
        enumerable: false,
        configurable: true
    });
    DataGridSortIndicatorDescription.$t = markType(DataGridSortIndicatorDescription, 'DataGridSortIndicatorDescription', Description.$);
    return DataGridSortIndicatorDescription;
}(Description));
export { DataGridSortIndicatorDescription };
