// igniteui-react-core/esm5/lib/DataGridFilterDialogDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataGridFilterDialogDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataGridFilterDialogDescription, _super);
    function DataGridFilterDialogDescription() {
        var _this = _super.call(this) || this;
        _this.h = false;
        _this.k = 0;
        _this.j = 0;
        _this.t = null;
        _this.l = 0;
        _this.p = null;
        _this.u = null;
        _this.q = null;
        _this.s = null;
        _this.r = null;
        _this.v = null;
        return _this;
    }
    DataGridFilterDialogDescription.prototype.get_type = function () {
        return "DataGridFilterDialog";
    };
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "isAnimationEnabled", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("IsAnimationEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("PixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "actualPixelScalingRatio", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ActualPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "iconColor", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("IconColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "viewSize", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ViewSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "commandCompletedRef", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("CommandCompletedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "invalidateActionsRef", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("InvalidateActionsRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "dialogOpeningRef", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("DialogOpeningRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "filterChangingRef", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("FilterChangingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "filterChangedRef", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("FilterChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridFilterDialogDescription.prototype, "renderCompletedRef", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("RenderCompletedRef");
        },
        enumerable: false,
        configurable: true
    });
    DataGridFilterDialogDescription.$t = markType(DataGridFilterDialogDescription, 'DataGridFilterDialogDescription', Description.$);
    return DataGridFilterDialogDescription;
}(Description));
export { DataGridFilterDialogDescription };
