// igniteui-react-core/esm5/lib/DataLegendHeaderDateMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Describes mode for formatting date in header of data legend
 */
export var DataLegendHeaderDateMode = /*@__PURE__*/ (function (DataLegendHeaderDateMode) {
    /**
     * Specifies automatic date format that defaults to short string
     */
    DataLegendHeaderDateMode[DataLegendHeaderDateMode["Auto"] = 0] = "Auto";
    /**
     * Specifies date as short string in header row of data legend
     */
    DataLegendHeaderDateMode[DataLegendHeaderDateMode["ShortStyle"] = 1] = "ShortStyle";
    /**
     * Specifies date as long string in header row of data legend
     */
    DataLegendHeaderDateMode[DataLegendHeaderDateMode["LongStyle"] = 2] = "LongStyle";
    /**
     * Specifies date as medium string in header row of data legend
     */
    DataLegendHeaderDateMode[DataLegendHeaderDateMode["MediumStyle"] = 3] = "MediumStyle";
    /**
     * Specifies date as full string in header row of data legend
     */
    DataLegendHeaderDateMode[DataLegendHeaderDateMode["FullStyle"] = 4] = "FullStyle";
    /**
     * Specifies no date in header row of data legend
     */
    DataLegendHeaderDateMode[DataLegendHeaderDateMode["None"] = 5] = "None";
    return DataLegendHeaderDateMode;
})({});
/**
 * @hidden
 */
export var DataLegendHeaderDateMode_$type = /*@__PURE__*/ markEnum('DataLegendHeaderDateMode', 'Auto,0|ShortStyle,1|LongStyle,2|MediumStyle,3|FullStyle,4|None,5');
