// igniteui-react-core/esm5/lib/DataVisualizationLocaleEn.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocaleEn = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocaleEn() {
        this["AssemblyNotIncluded"] = "Cannot load {0}.dll. You need to make sure that this assembly is included in the XAP that is deployed to the Windows Phone, for example, by adding a reference to it from any of your projects.";
        this["BreakEvenTitle_BreakEven"] = "Break Even";
        this["BreakEvenTitle_FixedCost"] = "Fixed Cost";
        this["BreakEvenTitle_LossArea"] = "Loss Area";
        this["BreakEvenTitle_MarginalProfit"] = "Marginal Profit";
        this["BreakEvenTitle_ProfitArea"] = "Profit Area";
        this["BreakEvenTitle_SafetyMargin"] = "Safety Margin";
        this["BreakEvenTitle_SalesRevenue"] = "Sales Revenue";
        this["BreakEvenTitle_TotalCost"] = "Total Cost";
        this["BreakEvenTitle_VariableCost"] = "Variable Cost";
        this["BubbleSeries_Radius"] = "Radius";
        this["DataChart_InteractivityNotLoaded"] = "Interactivity functionality was accessed but chart interaction module has not been loaded.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Number abbreviation functionality was used, but the number abbreviator module has not been loaded.";
        this["DataChart_VisualDataNotLoaded"] = "VisualData functionality was accessed but chart interaction module has not been loaded.";
        this["DataProviderNotInitialized"] = "Data provider was used without initializing first. Data provider must be initialized before use.";
        this["DataSource_Summary_Avg"] = "Avg";
        this["DataSource_Summary_Count"] = "Count";
        this["DataSource_Summary_Max"] = "Max";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Sum";
        this["Default_Series_Title"] = "Series Title";
        this["FinancialChart_IndicatorMenu_Header"] = "Indicators";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Standalone Indicators";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Overlays";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Trendlines";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volumes";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Absolute Volume Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Accumulation/Distribution";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Average Directional Index";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Average True Range";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Bollinger Band Width";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Chaikin Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Chaikin Volatility";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Commodity Channel Index";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Detrended Price Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Ease Of Movement";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Fast Stochastic Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Force Index";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Full Stochastic Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Market Facilitation Index";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Mass Index";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Median Price";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Money Flow Index";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Negative Volume Index";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "On Balance Volume";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Percentage Price Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Percentage Volume Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Positive Volume Index";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Price Volume Trend";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Slow Stochastic Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Standard Deviation";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Typical Price";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Ultimate Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Weighted Close";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bollinger Bands";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Price Channel";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Cubic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Cumulative Average";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Exponential Average";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Exponential Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Linear Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Logarithmic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Modified Average";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "PowerLaw Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Quadratic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Quartic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Quintic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Simple Average";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Weighted Average";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Area";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Column";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Line";
        this["FinancialChart_RangeSelector_All"] = "All";
        this["FinancialChart_RangeSelector_From"] = "From";
        this["FinancialChart_RangeSelector_OneMonth"] = "1M";
        this["FinancialChart_RangeSelector_OneYear"] = "1Y";
        this["FinancialChart_RangeSelector_SixMonths"] = "6M";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3M";
        this["FinancialChart_RangeSelector_To"] = "To";
        this["FinancialChart_RangeSelector_YearToDate"] = "YTD";
        this["FinancialSeries_Close"] = "Close";
        this["FinancialSeries_High"] = "High";
        this["FinancialSeries_Low"] = "Low";
        this["FinancialSeries_Open"] = "Open";
        this["FinancialSeries_Volume"] = "Volume";
        this["NoEncodingsLoaded"] = "No Encodings are found! Please load the Infragistics.Encoding assembly or look at the documentation how to create custom encodings.";
        this["NotSupportedEncoding"] = "The encoding corresponding under the {0} {1} is not supported.";
        this["Object_Sealed"] = "Modifications are not permitted after the object has been sealed.";
        this["OPD_DefaultInteraction"] = "Select the default mouse drag interaction";
        this["OPD_DragPan"] = "Pan the surface";
        this["OPD_DragZoom"] = "Zoom in to selection area";
        this["OPD_ScaleToFit"] = "Fit to Screen";
        this["OPD_ScaleToFit_SeriesViewer"] = "Reset Zoom (100%)";
        this["OPD_ZoomIn"] = "Zoom in";
        this["OPD_ZoomOut"] = "Zoom out";
        this["OPD_ZoomTo100"] = "Zoom to 100%";
        this["PieChart_Others"] = "Others";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Range based modifications of the data source are not currently supported";
        this["ScatterSeries_Value"] = "Value";
        this["TRIAL_VERSION"] = "UNLICENSED TRIAL Buy at Infragistics.com";
    }
    return DataVisualizationLocaleEn;
}());
export { DataVisualizationLocaleEn };
