// igniteui-react-core/esm5/lib/DataContext.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, INotifyPropertyChanged_$type, PropertyChangedEventArgs, markType } from "./type";
import { ILegendContext_$type } from "./ILegendContext";
/**
 * @hidden
 */
var DataContext = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataContext, _super);
    function DataContext() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.c = 0;
        _this.g = null;
        _this.d = null;
        _this.j = null;
        _this.l = null;
        _this.e = null;
        _this.k = null;
        _this.a = 0;
        _this.f = null;
        _this.m = null;
        _this.propertyChanged = null;
        return _this;
    }
    Object.defineProperty(DataContext.prototype, "series", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "item", {
        get: function () {
            return this.d;
        },
        set: function (a) {
            if (a != this.d) {
                this.d = a;
                this.i("Item");
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "actualItemBrush", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            if (a != this.j) {
                this.j = a;
                this.i("ActualItemBrush");
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "outline", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            if (a != this.l) {
                this.l = a;
                this.i("Outline");
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "itemLabel", {
        get: function () {
            return this.e;
        },
        set: function (a) {
            if (a != this.e) {
                this.e = a;
                this.i("ItemLabel");
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "itemBrush", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            if (a != this.k) {
                this.k = a;
                this.i("ItemBrush");
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "thickness", {
        get: function () {
            return this.a;
        },
        set: function (a) {
            if (a != this.a) {
                this.a = a;
                this.i("Thickness");
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "legendLabel", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            if (a != this.f) {
                this.f = a;
                this.i("LegendLabel");
            }
        },
        enumerable: false,
        configurable: true
    });
    DataContext.prototype.i = function (a) {
        if (this.propertyChanged != null) {
            this.propertyChanged(this, new PropertyChangedEventArgs(a));
        }
    };
    DataContext.prototype.flatten = function () {
        var series_ = this.series;
        var item_ = this.item;
        var actualItemBrush_ = this.actualItemBrush;
        var outline_ = this.outline;
        var itemLabel_ = this.itemLabel;
        var itemBrush_ = this.itemBrush;
        var thickness_ = this.thickness;
        return { item: item_, itemBrush: actualItemBrush_, outline: outline_, itemLabel: itemLabel_, thickness: thickness_ };
    };
    DataContext.$t = markType(DataContext, 'DataContext', Base.$, [ILegendContext_$type, INotifyPropertyChanged_$type]);
    return DataContext;
}(Base));
export { DataContext };
