// igniteui-react-core/esm5/lib/DataVisualizationLocaleTr.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocaleTr = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocaleTr() {
        this["AssemblyNotIncluded"] = "{0}.dll yüklenemiyor. Bu derlemenin Windows Phone'a dağıtılan XAP'ye dahil edildiğinden emin olmanız gerekir, örneğin, projelerinizden herhangi birinden ona bir referans ekleyerek.";
        this["BreakEvenTitle_BreakEven"] = "Maliyeti Karşılayan";
        this["BreakEvenTitle_FixedCost"] = "Sabit Maliyet";
        this["BreakEvenTitle_LossArea"] = "Zarar Alanı";
        this["BreakEvenTitle_MarginalProfit"] = "Marjinal Kâr";
        this["BreakEvenTitle_ProfitArea"] = "Kâr Alanı";
        this["BreakEvenTitle_SafetyMargin"] = "Güvenlik Marjı";
        this["BreakEvenTitle_SalesRevenue"] = "Satış Geliri";
        this["BreakEvenTitle_TotalCost"] = "Toplam Maliyet";
        this["BreakEvenTitle_VariableCost"] = "Değişken Maliyet";
        this["BubbleSeries_Radius"] = "Yarıçap";
        this["DataChart_InteractivityNotLoaded"] = "Etkileşim işlevine erişildi, ancak grafik etkileşim modülü yüklenmedi.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Sayı kısaltma işlevi kullanıldı, ancak sayı kısaltma modülü yüklenmedi.";
        this["DataChart_VisualDataNotLoaded"] = "VisualData işlevine erişildi, ancak grafik etkileşim modülü yüklenmedi.";
        this["DataProviderNotInitialized"] = "Veri sağlayıcı, önce başlatılmadan kullanıldı. Veri sağlayıcı, kullanımdan önce başlatılmalıdır.";
        this["DataSource_Summary_Avg"] = "Ort.";
        this["DataSource_Summary_Count"] = "Adet";
        this["DataSource_Summary_Max"] = "Maks";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Toplam";
        this["Default_Series_Title"] = "Dizi Başlığı";
        this["FinancialChart_IndicatorMenu_Header"] = "Göstergeler";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Bağımsız Göstergeler";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Yer paylaşımları";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Trend çizgileri";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Ciltler";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Mutlak Hacim Osilatörü";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Birikim/Dağılımı";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Ortalama Yön Endeksi";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Ortalama Doğru Aralığı";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Bollinger Bant Genişliği";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Chaikin Osilatör";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Chaikin Volatilitesi";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Emtia Kanal Endeksi";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Azalan Fiyat Osilatörü";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Hareket Kolaylığı";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Hızlı Stokastik Osilatör";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Kuvvet Endeksi";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Tam Stokastik Osilatör";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Piyasa Kolaylaştırma Endeksi";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Kütle Endeksi";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Medyan Fiyat";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Para Akışı Endeksi";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Negatif Hacim Endeksi";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "Denge İşlem Hacmi";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Yüzde Fiyat Osilatörü";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Yüzde Hacim Osilatörü";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Pozitif Hacim Endeksi";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Fiyat Hacim Eğilimi";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Yavaş Stokastik Osilatör";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Standart Sapma";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Tipik Fiyat";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Nihai Osilatör";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Ağırlıklı Kapanış";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams% R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bollinger bantları";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Fiyat Kanalı";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Kübik Uyum";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Kümülatif Ortalama";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Üssel Ortalama";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Üssel Uyum";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Doğrusal Uyum";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Logaritmik Uyum";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Değiştirilmiş Ortalama";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "Kuvvet Uyumu";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "İkinci Dereceden Uyum";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Dördüncü Dereceden Uyum";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Beşinci Dereceden Uyum";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Basit Ortalama";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Ağırlıklı Ortalama";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Alan";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Sütun";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Çizgi";
        this["FinancialChart_RangeSelector_All"] = "Tümü";
        this["FinancialChart_RangeSelector_From"] = "Kaynak";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 Ay";
        this["FinancialChart_RangeSelector_OneYear"] = "1 yıl";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 Ay";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 Ay";
        this["FinancialChart_RangeSelector_To"] = "Hedef";
        this["FinancialChart_RangeSelector_YearToDate"] = "Yıl Başından Beri";
        this["FinancialSeries_Close"] = "Kapat";
        this["FinancialSeries_High"] = "Yüksek";
        this["FinancialSeries_Low"] = "Düşük";
        this["FinancialSeries_Open"] = "Açık";
        this["FinancialSeries_Volume"] = "Ses";
        this["NoEncodingsLoaded"] = "Kodlama bulunamadı! Lütfen Infragistics.Encoding derlemesini yükleyin veya özel kodlamaların nasıl oluşturulacağına ilişkin belgelere bakın.";
        this["NotSupportedEncoding"] = "{0} {1} altında karşılık gelen kodlama desteklenmiyor.";
        this["Object_Sealed"] = "Nesne mühürlendikten sonra değişikliklere izin verilmez.";
        this["OPD_DefaultInteraction"] = "Varsayılan fare sürükleme etkileşimini seçin";
        this["OPD_DragPan"] = "Yüzeyi kaydırın";
        this["OPD_DragZoom"] = "Seçim alanına yakınlaştırın";
        this["OPD_ScaleToFit"] = "Ekrana Sığdır";
        this["OPD_ScaleToFit_SeriesViewer"] = "Yakınlaştırmayı Sıfırla (%100)";
        this["OPD_ZoomIn"] = "Yakınlaştır";
        this["OPD_ZoomOut"] = "Uzaklaştır";
        this["OPD_ZoomTo100"] = "%100 Yakınlaştır";
        this["PieChart_Others"] = "Diğerleri";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Veri kaynağının aralığa dayalı değişiklikleri şu anda desteklenmemektedir";
        this["ScatterSeries_Value"] = "Değer";
        this["TRIAL_VERSION"] = "LİSANSSIZ DENEME Infragistics.com'dan Satın Alın";
    }
    return DataVisualizationLocaleTr;
}());
export { DataVisualizationLocaleTr };
