// igniteui-react-core/esm5/lib/DataChartMouseButtonEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataChartMouseButtonEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataChartMouseButtonEventArgsDescription, _super);
    function DataChartMouseButtonEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.o = false;
        _this.n = false;
        _this.r = null;
        _this.j = null;
        _this.h = null;
        _this.l = null;
        return _this;
    }
    DataChartMouseButtonEventArgsDescription.prototype.get_type = function () {
        return "DataChartMouseButtonEventArgs";
    };
    Object.defineProperty(DataChartMouseButtonEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataChartMouseButtonEventArgsDescription.prototype, "handled", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Handled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataChartMouseButtonEventArgsDescription.prototype, "cancelSelection", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("CancelSelection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataChartMouseButtonEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataChartMouseButtonEventArgsDescription.prototype, "series", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Series");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataChartMouseButtonEventArgsDescription.prototype, "plotAreaPosition", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("PlotAreaPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataChartMouseButtonEventArgsDescription.prototype, "chart", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Chart");
        },
        enumerable: false,
        configurable: true
    });
    DataChartMouseButtonEventArgsDescription.$t = markType(DataChartMouseButtonEventArgsDescription, 'DataChartMouseButtonEventArgsDescription', Description.$);
    return DataChartMouseButtonEventArgsDescription;
}(Description));
export { DataChartMouseButtonEventArgsDescription };
