// igniteui-react-core/esm5/lib/DataLegendSummaryEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataLegendSummaryEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataLegendSummaryEventArgsDescription, _super);
    function DataLegendSummaryEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.l = null;
        _this.j = 0;
        _this.n = null;
        _this.m = null;
        return _this;
    }
    DataLegendSummaryEventArgsDescription.prototype.get_type = function () {
        return "DataLegendSummaryEventArgs";
    };
    Object.defineProperty(DataLegendSummaryEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendSummaryEventArgsDescription.prototype, "columnValues", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ColumnValues");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendSummaryEventArgsDescription.prototype, "columnMemberPath", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ColumnMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendSummaryEventArgsDescription.prototype, "summaryValue", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("SummaryValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendSummaryEventArgsDescription.prototype, "summaryUnits", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("SummaryUnits");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataLegendSummaryEventArgsDescription.prototype, "summaryLabel", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("SummaryLabel");
        },
        enumerable: false,
        configurable: true
    });
    DataLegendSummaryEventArgsDescription.$t = markType(DataLegendSummaryEventArgsDescription, 'DataLegendSummaryEventArgsDescription', Description.$);
    return DataLegendSummaryEventArgsDescription;
}(Description));
export { DataLegendSummaryEventArgsDescription };
