// igniteui-react-core/esm5/lib/DataVisualizationLocaleDe.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocaleDe = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocaleDe() {
        this["AssemblyNotIncluded"] = "{0}.dll kann nicht geladen werden. Stellen Sie sicher, dass diese Assembly im für Windows Phone bereitgestellten XAP enthalten ist; fügen Sie ihr hierzu beispielsweise einen Verweis von einem Ihrer Projekte hinzu.";
        this["BreakEvenTitle_BreakEven"] = "Break-Even";
        this["BreakEvenTitle_FixedCost"] = "Fixkosten";
        this["BreakEvenTitle_LossArea"] = "Verlustbereich";
        this["BreakEvenTitle_MarginalProfit"] = "Grenzertrag";
        this["BreakEvenTitle_ProfitArea"] = "Gewinnbereich";
        this["BreakEvenTitle_SafetyMargin"] = "Sicherheitsmarge";
        this["BreakEvenTitle_SalesRevenue"] = "Umsatzerlös";
        this["BreakEvenTitle_TotalCost"] = "Gesamtkosten";
        this["BreakEvenTitle_VariableCost"] = "Variable Kosten";
        this["BubbleSeries_Radius"] = "Radius";
        this["DataChart_InteractivityNotLoaded"] = "Es wurde auf die Interaktivitätsfunktion zugegriffen, aber das Modul für die Diagramminteraktion wurde nicht geladen.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Number abbreviation functionality was used, but the number abbreviator module has not been loaded.";
        this["DataChart_VisualDataNotLoaded"] = "Es wurde auf die VisualData-Funktion zugegriffen, aber das Modul für die Diagramminteraktion wurde nicht geladen.";
        this["DataProviderNotInitialized"] = "Der Datenanbieter wurde ohne vorherige Initialisierung verwendet. Der Datenanbieter muss vor der Verwendung initialisiert werden.";
        this["DataSource_Summary_Avg"] = "Avg";
        this["DataSource_Summary_Count"] = "Count";
        this["DataSource_Summary_Max"] = "Max";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Sum";
        this["Default_Series_Title"] = "Reihentitel";
        this["FinancialChart_IndicatorMenu_Header"] = "Anzeigen";
        this["FinancialChart_IndicatorMenu_Header1"] = "Anzeigen";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Eigenständige Anzeigen";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader1"] = "Eigenständige Anzeigen";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Überlagerungen";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader1"] = "Überlagerungen";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Trendlinien";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader1"] = "Trendlinien";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volumen";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader1"] = "Volumen";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Oszillator absolute Menge";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Akkumulationsverteilung";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Durchschnittlicher Richtungsindex";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Durchschnittlicher wahrer Bereich";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Breite des Bollinger-Bands";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Chaikin-Oszillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Chaikin-Volatilität";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Commodity Channel Index";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Oszillator trendbereinigter Preis";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Bewegungsleichtigkeit";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Schneller stochastischer Oszillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Kraftindex";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Vollständiger stochastischer Oszillator";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Market Facilitation Index";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Massenindex";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Durchschnittlicher Preis";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Geldflussindex";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Negativer Volumenindex";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "On Balance Volume";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Percentage Price Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Percentage Volume Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Positive Volume Index";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "On-Balance-Volumen";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Langsamer stochastischer Oszillator";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Standardabweichung";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stochastik Relative-Stärke-Index";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Typischer Preis";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Ultimate Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Gewichtetes Mittel";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams-Prozentbereich";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bollinger-Bänder";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Preiskanal";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Kubische Anpassung";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Kumulativer Durchschnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Exponentieller Durchschnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Exponentielle Anpassung";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Lineare Anpassung";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Logarithmische Anpassung";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Modifizierter Durchschnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "Potenzgesetz-Entsprechung";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Quadratische Anpassung";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Anpassung vierten Grades";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Anpassung fünften Grades";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Einfacher Durchschnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Gewichteter Durchschnitt";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Bereich";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Spalte";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Linie";
        this["FinancialChart_RangeSelector_All"] = "Alle";
        this["FinancialChart_RangeSelector_From"] = "From";
        this["FinancialChart_RangeSelector_OneMonth"] = "1M";
        this["FinancialChart_RangeSelector_OneYear"] = "1Y";
        this["FinancialChart_RangeSelector_SixMonths"] = "6M";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3M";
        this["FinancialChart_RangeSelector_To"] = "To";
        this["FinancialChart_RangeSelector_YearToDate"] = "YTD";
        this["FinancialSeries_Close"] = "Schließen";
        this["FinancialSeries_High"] = "Hoch";
        this["FinancialSeries_Low"] = "Niedrig";
        this["FinancialSeries_Open"] = "Öffnen";
        this["FinancialSeries_Volume"] = "Lautstärke";
        this["NoEncodingsLoaded"] = "Es wurden keine Codierungen gefunden! Laden Sie die Infragistics.Encoding Assembly oder lesen Sie in der Dokumentation nach, wie Sie benutzerdefinierte Codierungen erstellen.";
        this["NotSupportedEncoding"] = "Die Codierung entsprechend {0} {1} wird nicht unterstützt.";
        this["Object_Sealed"] = "Modifications are not permitted after the object has been sealed.";
        this["OPD_DefaultInteraction"] = "Standardmäßige Interaktion für Ziehen mit Maustaste auswählen";
        this["OPD_DragPan"] = "Die Oberfläche schwenken";
        this["OPD_DragZoom"] = "Auswahlbereich vergrößern";
        this["OPD_ScaleToFit"] = "An Bildschirm anpassen";
        this["OPD_ScaleToFit_SeriesViewer"] = "Zoom zurücksetzen (100%)";
        this["OPD_ZoomIn"] = "Vergrößern";
        this["OPD_ZoomOut"] = "Verkleinern";
        this["OPD_ZoomTo100"] = "Zoommodus 100%";
        this["PieChart_Others"] = "Andere";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Bereichsbasierte Änderungen an der Datenquelle werden aktuell nicht unterstützt";
        this["ScatterSeries_Value"] = "Wert";
        this["TRIAL_VERSION"] = "NICHT LIZENZIERTER TEST Zu kaufen auf Infragistics.com";
    }
    return DataVisualizationLocaleDe;
}());
export { DataVisualizationLocaleDe };
