// igniteui-react-core/esm5/lib/DataVisualizationLocaleIt.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocaleIt = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocaleIt() {
        this["AssemblyNotIncluded"] = "Impossibile caricare {0}.dll. È necessario assicurarsi che questo assembly sia incluso nell’XAP distribuito su Windows Phone, ad esempio aggiungendo un riferimento ad esso da uno qualsiasi dei progetti.";
        this["BreakEvenTitle_BreakEven"] = "Pareggio finanziario";
        this["BreakEvenTitle_FixedCost"] = "Costo fisso";
        this["BreakEvenTitle_LossArea"] = "Area perdite";
        this["BreakEvenTitle_MarginalProfit"] = "Profitto marginale";
        this["BreakEvenTitle_ProfitArea"] = "Area di profitto";
        this["BreakEvenTitle_SafetyMargin"] = "Margine di sicurezza";
        this["BreakEvenTitle_SalesRevenue"] = "Entrate di vendita";
        this["BreakEvenTitle_TotalCost"] = "Costo totale";
        this["BreakEvenTitle_VariableCost"] = "Costo variabile";
        this["BubbleSeries_Radius"] = "Raggio";
        this["DataChart_InteractivityNotLoaded"] = "Accesso alla funzionalità di interattività ma il modulo di interazione grafico non è stato caricato.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "È stata utilizzata la funzionalità di abbreviazione numerica, ma il modulo di abbreviazione numerica non è stato caricato.";
        this["DataChart_VisualDataNotLoaded"] = "Accesso alla funzionalità VisualData ma il modulo di interazione grafico non è stato caricato.";
        this["DataProviderNotInitialized"] = "Il provider di dati è stato utilizzato senza prima inizializzare. Il provider di dati deve essere inizializzato prima dell'uso.";
        this["DataSource_Summary_Avg"] = "Media";
        this["DataSource_Summary_Count"] = "Conteggio";
        this["DataSource_Summary_Max"] = "Max";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Somma";
        this["Default_Series_Title"] = "Titolo serie";
        this["FinancialChart_IndicatorMenu_Header"] = "Indicatori";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Indicatori autonomi";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Sovrapposizioni";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Linee di tendenza";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volumi";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Oscillatore di volume assoluto";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Distribuzione accumulo";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Indice direzionale medio";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Intervallo medio vero";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Larghezza della banda di Bollinger";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Chaikin Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Volatilità del Chaikin";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Indice dei canali merceologici";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Oscillatore a prezzo differenziato";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Facilità di movimento";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Oscillatore stocastico veloce";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Force Index";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Oscillatore stocastico completo";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Indice di facilitazione del mercato";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Indice di massa";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Prezzo medio";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Indice del flusso di denaro";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "Convergenza/divergenza media mobile";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Indice di volume negativo";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "Volume del saldo";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Oscillatore prezzo percentuale";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Oscillatore volume percentuale";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Indice di volume positivo";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Andamento volume prezzo";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "Indice di forza relativa (RSI)";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Oscillatore stocastico lento";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Deviazione standard";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Prezzo tipico";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Oscillatore Ultimate";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Chiusura ponderata";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bande di Bollinger";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Canale di prezzo";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Adattamento a terzo grado";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Media cumulativa";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Media esponenziale";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Adattamento esponenziale";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Adattamento lineare";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Adattamento logaritmico";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Media modificata";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "PowerLaw Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Adattamento a secondo grado";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Adattamento a quarto grado";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Adattamento a quinto grado";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Media semplice";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Media ponderata";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Area";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Colonna";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Riga";
        this["FinancialChart_RangeSelector_All"] = "Tutto";
        this["FinancialChart_RangeSelector_From"] = "Da";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 mese";
        this["FinancialChart_RangeSelector_OneYear"] = "1 anno";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 mesi";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 mesi";
        this["FinancialChart_RangeSelector_To"] = "A";
        this["FinancialChart_RangeSelector_YearToDate"] = "Da inizio anno";
        this["FinancialSeries_Close"] = "Chiudi";
        this["FinancialSeries_High"] = "Alto";
        this["FinancialSeries_Low"] = "Basso";
        this["FinancialSeries_Open"] = "Aperto";
        this["FinancialSeries_Volume"] = "Volume";
        this["NoEncodingsLoaded"] = "Nessuna codifica trovata! Caricare l’assembly Infragistics.Encoding o consultare la documentazione su come creare codifiche personalizzate.";
        this["NotSupportedEncoding"] = "La codifica corrispondente in {0} {1} non è supportata.";
        this["Object_Sealed"] = "Modifiche non sono consentite dopo che l'oggetto è stato sigillato.";
        this["OPD_DefaultInteraction"] = "Selezionare l’interazione di trascinamento del mouse predefinita";
        this["OPD_DragPan"] = "Panoramica della superficie";
        this["OPD_DragZoom"] = "Zoom l'area di selezione";
        this["OPD_ScaleToFit"] = "Adatta allo schermo";
        this["OPD_ScaleToFit_SeriesViewer"] = "Reimposta zoom (100%)";
        this["OPD_ZoomIn"] = "Zoom avanti";
        this["OPD_ZoomOut"] = "Zoom indietro";
        this["OPD_ZoomTo100"] = "Zoom 100%";
        this["PieChart_Others"] = "Altri";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Modifiche dell’origine dati basate sull’intervallo non sono attualmente supportate";
        this["ScatterSeries_Value"] = "Valore";
        this["TRIAL_VERSION"] = "PROVA SENZA LICENZA Per acquistare: Infragistics.com";
    }
    return DataVisualizationLocaleIt;
}());
export { DataVisualizationLocaleIt };
