// igniteui-react-core/esm5/lib/DataVisualizationLocaleCs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocaleCs = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocaleCs() {
        this["AssemblyNotIncluded"] = "Nelze načíst soubor {0}.dll. Musíte se ujistit, že toto sestavení je zahrnuto v XAP, který je nasazen na Windows Phone, například přidáním odkazu na něj z některého z vašich projektů.";
        this["BreakEvenTitle_BreakEven"] = "Beze ztrát";
        this["BreakEvenTitle_FixedCost"] = "Fixní cena";
        this["BreakEvenTitle_LossArea"] = "Ztráta prostoru";
        this["BreakEvenTitle_MarginalProfit"] = "Mezní zisk";
        this["BreakEvenTitle_ProfitArea"] = "Zisková oblast";
        this["BreakEvenTitle_SafetyMargin"] = "Bezpečnostní marže";
        this["BreakEvenTitle_SalesRevenue"] = "Tržby";
        this["BreakEvenTitle_TotalCost"] = "Celkové náklady";
        this["BreakEvenTitle_VariableCost"] = "Variabilní náklady";
        this["BubbleSeries_Radius"] = "Poloměr";
        this["DataChart_InteractivityNotLoaded"] = "Byla aktivována funkce interaktivity, ale nebyl načten modul interakce s grafy.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Byla použita funkce zkratky čísel, ale modul zkratky čísel nebyl načten.";
        this["DataChart_VisualDataNotLoaded"] = "Byla zpřístupněna funkce VisualData, ale modul interakce s grafy nebyl načten.";
        this["DataProviderNotInitialized"] = "Poskytovatel dat byl použit bez první inicializace. Před použitím je nutné inicializovat poskytovatele dat.";
        this["DataSource_Summary_Avg"] = "Prům";
        this["DataSource_Summary_Count"] = "Počet";
        this["DataSource_Summary_Max"] = "Max";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Součet";
        this["Default_Series_Title"] = "Název série";
        this["FinancialChart_IndicatorMenu_Header"] = "Indikátory";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Samostatné ukazatele";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Překryvy";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Trendové čáry";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Svazky";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Oscilátor absolutního objemu";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Distribuce/Akumulace";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Průměrný směrový index";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Průměrný skutečný rozsah";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Šířka pásma Bollinger";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Chaikin oscilátor";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Chaikin Volatility";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Index komoditních kanálů";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Detrended Price Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Snadnost pohybu";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Rychlý stochastický oscilátor";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Index síly";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Plný stochastický oscilátor";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Index usnadnění trhu";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Hromadný index";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Střední cena";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Index peněžního toku";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Negativní objemový index";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "Objem zůstatku";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Procentní cenový oscilátor";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Procentní objemový oscilátor";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Pozitivní index objemu";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Trend objemu cen";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Pomalý stochastický oscilátor";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Standardní odchylka";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Typická cena";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Konečný oscilátor";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Vážený blízko";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bollingerovy kapely";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Cenový kanál";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Cubic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Kumulativní průměr";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Exponenciální průměr";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Exponenciální přizpůsobení";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Lineární uložení";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Logaritmické přizpůsobení";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Upravený průměr";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "PowerLaw Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Kvadratický tvar";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Quartic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Quintic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Jednoduchý průměr";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Vážený průměr";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Plocha";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Sloupec";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Čára";
        this["FinancialChart_RangeSelector_All"] = "Všechno";
        this["FinancialChart_RangeSelector_From"] = "Z";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 měsíc";
        this["FinancialChart_RangeSelector_OneYear"] = "1 rok";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 měsíce";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 měsíce";
        this["FinancialChart_RangeSelector_To"] = "Na";
        this["FinancialChart_RangeSelector_YearToDate"] = "YTD (Rok do data)";
        this["FinancialSeries_Close"] = "Zavřít";
        this["FinancialSeries_High"] = "Vysoký";
        this["FinancialSeries_Low"] = "Nízký";
        this["FinancialSeries_Open"] = "Otevřeno";
        this["FinancialSeries_Volume"] = "Objem";
        this["NoEncodingsLoaded"] = "Nebyly nalezeny žádné kódování! Načtěte sestavu Infragistics.Encoding nebo se podívejte do dokumentace, jak vytvořit vlastní kódování.";
        this["NotSupportedEncoding"] = "Kódování odpovídající pod {0} {1} není podporováno.";
        this["Object_Sealed"] = "Po zapečetění objektu nejsou úpravy povoleny.";
        this["OPD_DefaultInteraction"] = "Vyberte výchozí interakci tažení myší";
        this["OPD_DragPan"] = "Posuňte povrch";
        this["OPD_DragZoom"] = "Přibližte oblast výběru";
        this["OPD_ScaleToFit"] = "Přizpůsobit obrazovce";
        this["OPD_ScaleToFit_SeriesViewer"] = "Obnovit přiblížení (100%)";
        this["OPD_ZoomIn"] = "Přiblížit";
        this["OPD_ZoomOut"] = "Oddálit";
        this["OPD_ZoomTo100"] = "Přiblížit na 100%";
        this["PieChart_Others"] = "Ostatní";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Úpravy zdroje dat na základě rozsahu nejsou aktuálně podporovány";
        this["ScatterSeries_Value"] = "Hodnota";
        this["TRIAL_VERSION"] = "BEZPLATNÁ ZKUŠEBNÍ VERZE Nakupte na Infragistics.com";
    }
    return DataVisualizationLocaleCs;
}());
export { DataVisualizationLocaleCs };
