// igniteui-react-core/esm5/lib/DataVisualizationLocaleNb.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocaleNb = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocaleNb() {
        this["AssemblyNotIncluded"] = "Kan ikke laste inn {0}.dll. Du må sørge for at denne enheten er inkludert i XAP som er distribuert til Windows Phone, for eksempel ved å legge til en referanse til den fra et av prosjektene dine.";
        this["BreakEvenTitle_BreakEven"] = "Balanse";
        this["BreakEvenTitle_FixedCost"] = "Faste kostnader";
        this["BreakEvenTitle_LossArea"] = "Tapsområde";
        this["BreakEvenTitle_MarginalProfit"] = "Marginalfortjeneste";
        this["BreakEvenTitle_ProfitArea"] = "Profittområde";
        this["BreakEvenTitle_SafetyMargin"] = "Sikkerhetsmargin";
        this["BreakEvenTitle_SalesRevenue"] = "Salgsinntekter";
        this["BreakEvenTitle_TotalCost"] = "Totalkostnad";
        this["BreakEvenTitle_VariableCost"] = "Variable kostnader";
        this["BubbleSeries_Radius"] = "Radius";
        this["DataChart_InteractivityNotLoaded"] = "Interaktivitetsfunksjonalitet ble åpnet, men kartinteraksjonsmodulen er ikke lastet inn.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Nummerforkortelsesfunksjonalitet ble brukt, men nummerforkortelsesmodulen er ikke lastet inn.";
        this["DataChart_VisualDataNotLoaded"] = "VisualData-funksjonalitet ble åpnet, men kartinteraksjonsmodulen er ikke lastet inn.";
        this["DataProviderNotInitialized"] = "Dataleverandøren ble brukt uten å initialisere først. Dataleverandøren må initialiseres før bruk.";
        this["DataSource_Summary_Avg"] = "Gj.sn";
        this["DataSource_Summary_Count"] = "Telle";
        this["DataSource_Summary_Max"] = "Maks";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Sum";
        this["Default_Series_Title"] = "Serietittel";
        this["FinancialChart_IndicatorMenu_Header"] = "Indikatorer";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Frittstående indikatorer";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Overlegg";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Trendlinjer";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volumer";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Absolutt Volumoscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Akkumuleringsfordeling";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Gjennomsnittlig retningsindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Gjennomsnittlig True Range";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Bollinger bandbredde";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Chaikin Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Chaikin volatilitet";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Varekanalindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Detrended Price Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Lett å bevege seg";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Rask stokastisk oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Kraftindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Full stokastisk oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Markedsfasiliteringsindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Masseindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Median pris";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Pengestrømindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Negativ volumindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "On Balance Volume";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Prosentpris oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Prosent volumoscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Positiv volumindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Prisvolumtrend";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Langsom stokastisk oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Standardavvik";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Typisk pris";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Ultimate Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Vektet Lukk";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bollinger-band";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Priskanal";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Cubic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Kumulativ gjennomsnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Eksponensielt gjennomsnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Eksponensiell passform";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Lineær passform";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Logaritmisk passform";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Endret gjennomsnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "PowerLaw Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Kvadratisk passform";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Quartic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Quintic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Enkelt gjennomsnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Vektlagt gjennomsnitt";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Område";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Kolonne";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Linje";
        this["FinancialChart_RangeSelector_All"] = "Alle";
        this["FinancialChart_RangeSelector_From"] = "Fra";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 måned";
        this["FinancialChart_RangeSelector_OneYear"] = "1 år";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 måneder";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 måneder";
        this["FinancialChart_RangeSelector_To"] = "Til";
        this["FinancialChart_RangeSelector_YearToDate"] = "År til dato";
        this["FinancialSeries_Close"] = "Lukk";
        this["FinancialSeries_High"] = "Høy";
        this["FinancialSeries_Low"] = "Lav";
        this["FinancialSeries_Open"] = "Åpen";
        this["FinancialSeries_Volume"] = "Volum";
        this["NoEncodingsLoaded"] = "Ingen kodinger er funnet! Last inn Infragistics.Encoding-enheten, eller se på dokumentasjonen hvordan du lager egendefinerte kodinger.";
        this["NotSupportedEncoding"] = "Kodingen som tilsvarer under {0} {1} støttes ikke.";
        this["Object_Sealed"] = "Endringer er ikke tillatt etter at objektet er forseglet.";
        this["OPD_DefaultInteraction"] = "Velg standard musedrag-interaksjon";
        this["OPD_DragPan"] = "Panorer overflaten";
        this["OPD_DragZoom"] = "Zoom inn til valgområdet";
        this["OPD_ScaleToFit"] = "Tilpass til skjermen";
        this["OPD_ScaleToFit_SeriesViewer"] = "Tilbakestill zoom (100%)";
        this["OPD_ZoomIn"] = "Zoom inn";
        this["OPD_ZoomOut"] = "Zoome ut";
        this["OPD_ZoomTo100"] = "Zoom til 100%";
        this["PieChart_Others"] = "Andre";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Områdebaserte modifikasjoner av datakilden støttes foreløpig ikke";
        this["ScatterSeries_Value"] = "Verdi";
        this["TRIAL_VERSION"] = "ULISENSISERT PRØVE Kjøp på Infragistics.com";
    }
    return DataVisualizationLocaleNb;
}());
export { DataVisualizationLocaleNb };
