// igniteui-react-core/esm5/lib/DataVisualizationLocaleRo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocaleRo = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocaleRo() {
        this["AssemblyNotIncluded"] = "Nu se poate încărca {0}.dll. Trebuie să vă asigurați că acest ansamblu este inclus în XAP care este implementat pe Windows Phone, de exemplu, adăugând o referință la acesta din oricare dintre proiectele dvs.";
        this["BreakEvenTitle_BreakEven"] = "Pragul de rentabilitate";
        this["BreakEvenTitle_FixedCost"] = "Costul fix";
        this["BreakEvenTitle_LossArea"] = "Zona de pierdere";
        this["BreakEvenTitle_MarginalProfit"] = "Profit marginal";
        this["BreakEvenTitle_ProfitArea"] = "Zona de profit";
        this["BreakEvenTitle_SafetyMargin"] = "Marja de siguranță";
        this["BreakEvenTitle_SalesRevenue"] = "Veniturile din vânzări";
        this["BreakEvenTitle_TotalCost"] = "Cost total";
        this["BreakEvenTitle_VariableCost"] = "Costul variabil";
        this["BubbleSeries_Radius"] = "Rază";
        this["DataChart_InteractivityNotLoaded"] = "Funcționalitatea de interactivitate a fost accesată, dar modulul de interacțiune a diagramei nu a fost încărcat.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "A fost utilizată funcționalitatea de abreviere a numărului, dar modulul de abreviere a numărului nu a fost încărcat.";
        this["DataChart_VisualDataNotLoaded"] = "Funcționalitatea VisualData a fost accesată, dar modulul de interacțiune a diagramelor nu a fost încărcat.";
        this["DataProviderNotInitialized"] = "Furnizorul de date a fost utilizat fără a inițializa mai întâi. Furnizorul de date trebuie inițializat înainte de utilizare.";
        this["DataSource_Summary_Avg"] = "Media";
        this["DataSource_Summary_Count"] = "Numără";
        this["DataSource_Summary_Max"] = "Max";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Sumă";
        this["Default_Series_Title"] = "Titlul seriei";
        this["FinancialChart_IndicatorMenu_Header"] = "Indicatori";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Indicatori independenți";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Suprapuneri";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Linii de trend";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volumele";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Oscilator de volum absolut";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Distribuția/Acumulării";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Indicele direcțional mediu";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Interval adevărat mediu";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Lățimea benzii Bollinger";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Oscilator Chaikin";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Volatilitatea Chaikin";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Indicele canalului mărfurilor";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Oscilator de preț detras";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Ușurința mișcării";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Oscilator stocastic rapid";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Indice de forță";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Oscilator stocastic complet";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Indicele facilitării pieței";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Indicele de masă";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Preț mediu";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Indicele fluxului de bani";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Indicele volumului negativ";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "Volumul soldului";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Oscilator de preț procentual";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Oscilator de volum procentual";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Indicele volumului pozitiv";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Tendința volumului prețului";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Oscilator stocastic lent";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Deviație standard";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Preț tipic";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Oscilator final";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Închis ponderat";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Trupe Bollinger";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Canal Preț";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Potrivire cubică";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Media cumulativă";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Media exponențială";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Potrivire exponențială";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Linear Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Potrivire logaritmică";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Media modificată";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "PowerLaw Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Potrivire quadratică";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Potrivire cvartică";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Potrivire cvintică";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Media simplă";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Medie ponderată";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Zonă";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Coloană";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Linia";
        this["FinancialChart_RangeSelector_All"] = "Toate";
        this["FinancialChart_RangeSelector_From"] = "Din";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 lună";
        this["FinancialChart_RangeSelector_OneYear"] = "1 an";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 luni";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 luni";
        this["FinancialChart_RangeSelector_To"] = "La";
        this["FinancialChart_RangeSelector_YearToDate"] = "YTD";
        this["FinancialSeries_Close"] = "Închide";
        this["FinancialSeries_High"] = "Înalt";
        this["FinancialSeries_Low"] = "Scăzut";
        this["FinancialSeries_Open"] = "Deschis";
        this["FinancialSeries_Volume"] = "Volum";
        this["NoEncodingsLoaded"] = "Nu se găsesc codificări! Vă rugăm să încărcați ansamblul Infragistics.Encoding sau să consultați documentația despre cum să creați codificări personalizate.";
        this["NotSupportedEncoding"] = "Codificarea corespunzătoare sub {0} {1} nu este acceptată.";
        this["Object_Sealed"] = "Modificările nu sunt permise după ce obiectul a fost sigilat.";
        this["OPD_DefaultInteraction"] = "Selectați interacțiunea implicită de tragere a mouse-ului";
        this["OPD_DragPan"] = "Mișcați suprafața";
        this["OPD_DragZoom"] = "Măriți zona de selecție";
        this["OPD_ScaleToFit"] = "Încadrat in ecran";
        this["OPD_ScaleToFit_SeriesViewer"] = "Resetați zoomul (100%)";
        this["OPD_ZoomIn"] = "Mărește";
        this["OPD_ZoomOut"] = "Micșorează";
        this["OPD_ZoomTo100"] = "Măriți la 100%";
        this["PieChart_Others"] = "Alții";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Modificările sursei de date bazate pe intervale nu sunt acceptate în prezent";
        this["ScatterSeries_Value"] = "Valoare";
        this["TRIAL_VERSION"] = "VARIANTĂ DE PROBĂ FĂRĂ LICENȚĂ Cumpărați la Infragistics.com";
    }
    return DataVisualizationLocaleRo;
}());
export { DataVisualizationLocaleRo };
