// igniteui-react-core/esm5/lib/DataGridSummaryResultDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataGridSummaryResultDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataGridSummaryResultDescription, _super);
    function DataGridSummaryResultDescription() {
        var _this = _super.call(this) || this;
        _this.p = null;
        _this.m = null;
        _this.k = 0;
        _this.i = false;
        return _this;
    }
    DataGridSummaryResultDescription.prototype.get_type = function () {
        return "DataGridSummaryResult";
    };
    Object.defineProperty(DataGridSummaryResultDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSummaryResultDescription.prototype, "groupKeyRef", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("GroupKeyRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSummaryResultDescription.prototype, "value", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSummaryResultDescription.prototype, "summaryIndex", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("SummaryIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridSummaryResultDescription.prototype, "shouldDisplay", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ShouldDisplay");
        },
        enumerable: false,
        configurable: true
    });
    DataGridSummaryResultDescription.$t = markType(DataGridSummaryResultDescription, 'DataGridSummaryResultDescription', Description.$);
    DataGridSummaryResultDescription.__marshalByValue = true;
    DataGridSummaryResultDescription.__marshalByValueAlias = "DefaultSummaryResult";
    return DataGridSummaryResultDescription;
}(Description));
export { DataGridSummaryResultDescription };
