// igniteui-react-core/esm5/lib/DataVisualizationLocaleEs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocaleEs = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocaleEs() {
        this["AssemblyNotIncluded"] = "No se puede cargar {0}.dll. Asegúrese de que este conjunto esté incluido en el XAP que se ejecuta en el Windows Phone, por ejemplo añadiendo una referencia al mismo desde cualquiera de sus proyectos.";
        this["BreakEvenTitle_BreakEven"] = "Break Even";
        this["BreakEvenTitle_FixedCost"] = "Fixed Cost";
        this["BreakEvenTitle_LossArea"] = "Loss Area";
        this["BreakEvenTitle_MarginalProfit"] = "Marginal Profit";
        this["BreakEvenTitle_ProfitArea"] = "Profit Area";
        this["BreakEvenTitle_SafetyMargin"] = "Safety Margin";
        this["BreakEvenTitle_SalesRevenue"] = "Sales Revenue";
        this["BreakEvenTitle_TotalCost"] = "Total Cost";
        this["BreakEvenTitle_VariableCost"] = "Variable Cost";
        this["BubbleSeries_Radius"] = "Radio";
        this["DataChart_InteractivityNotLoaded"] = "Se ha accedido a la función de interactividad, pero no se ha cargado el módulo de interacción de gráficos.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Se ha utilizado la función de abreviatura de número, pero el módulo abreviador de número no se ha cargado.";
        this["DataChart_VisualDataNotLoaded"] = "Se ha accedido a la función Datos visuales, pero no se ha cargado el módulo de interacción de gráficos.";
        this["DataProviderNotInitialized"] = "El proveedor de datos se ha utilizado sin inicializarlo primero. El proveedor de datos debe inicializarse antes de utilizarlo.";
        this["DataSource_Summary_Avg"] = "Avg";
        this["DataSource_Summary_Count"] = "Count";
        this["DataSource_Summary_Max"] = "Max";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Sum";
        this["Default_Series_Title"] = "Título de la serie";
        this["FinancialChart_IndicatorMenu_Header"] = "Indicadores";
        this["FinancialChart_IndicatorMenu_Header1"] = "Indicadores";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Indicadores independientes";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader1"] = "Indicadores independientes";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Superposiciones";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader1"] = "Superposiciones";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Líneas de tendencia";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader1"] = "Líneas de tendencia";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volumen";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader1"] = "Volúmenes";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Oscilador de volumen absoluto";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Acumulación/Distribución";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Índice direccional promedio";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Rango medio verdadero";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Ancho de la Banda de Bollinger";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Oscilador de Chaikin";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Volatilidad de Chaikin";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Commodity Channel Index";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Oscilador de precio sin tendencia";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Facilidad de movimiento";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Oscilador estocástico rápido";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Force Index";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Oscilador estocástico total";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Índice de facilitación del mercado";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Índice de masa";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Precio medio";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Índice de flujo monetario";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Índice de volumen negativo";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "Balance de volúmenes";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Oscilador de precios en porcentaje";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Oscilador de volumen en porcentaje";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Índice de volumen positivo";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Price Volume Trend";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Oscilador estocástico lento";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Desviación estándar";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "RSI estocástico";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Precio típico";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Oscilador final";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Cierre ponderado";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bandas de Bollinger";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Canal de precios";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Ajuste cúbico";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Promedio acumulativo";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Promedio exponencial";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Ajuste exponencial";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Ajuste lineal";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Logarithmic Fit";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Ajuste logarítmico";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "Ajuste de ley de potencia";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Ajuste cuadrático";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Ajuste cuártico";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Ajuste quíntico";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Promedio simple";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Promedio ponderado";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Área";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Columna";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Línea";
        this["FinancialChart_RangeSelector_All"] = "Todo";
        this["FinancialChart_RangeSelector_From"] = "De";
        this["FinancialChart_RangeSelector_OneMonth"] = "1M";
        this["FinancialChart_RangeSelector_OneYear"] = "1Y";
        this["FinancialChart_RangeSelector_SixMonths"] = "6M";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3M";
        this["FinancialChart_RangeSelector_To"] = "A";
        this["FinancialChart_RangeSelector_YearToDate"] = "YTD";
        this["FinancialSeries_Close"] = "Cerrar";
        this["FinancialSeries_High"] = "Alto";
        this["FinancialSeries_Low"] = "Bajo";
        this["FinancialSeries_Open"] = "Abrir";
        this["FinancialSeries_Volume"] = "Volumen";
        this["NoEncodingsLoaded"] = "No se han encontrado codificaciones. Cargue el conjunto Infragistics.Encoding o consulte la documentación para saber cómo crear codificaciones personalizadas.";
        this["NotSupportedEncoding"] = "La codificación correspondiente bajo {0} {1} no se admite.";
        this["Object_Sealed"] = "No se permiten modificaciones después de haber sellado el objeto.";
        this["OPD_DefaultInteraction"] = "Seleccione la interacción predeterminada de arrastre del ratón";
        this["OPD_DragPan"] = "Aplicar panorámica a la superficie";
        this["OPD_DragZoom"] = "Zoom en el área de selección";
        this["OPD_ScaleToFit"] = "Ajustar a la pantalla";
        this["OPD_ScaleToFit_SeriesViewer"] = "Restablecer zoom (100%)";
        this["OPD_ZoomIn"] = "Acercar";
        this["OPD_ZoomOut"] = "Alejar";
        this["OPD_ZoomTo100"] = "Zoom al 100%";
        this["PieChart_Others"] = "Otros";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Las modificaciones de la fuente de datos basadas en intervalos no son compatibles";
        this["ScatterSeries_Value"] = "Valor";
        this["TRIAL_VERSION"] = "Compre la VERSIÓN DE PRUEBA SIN LICENCIA en Infragistics.com";
    }
    return DataVisualizationLocaleEs;
}());
export { DataVisualizationLocaleEs };
