// igniteui-react-core/esm5/lib/DataGridExpansionIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataGridExpansionIndicatorDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataGridExpansionIndicatorDescription, _super);
    function DataGridExpansionIndicatorDescription() {
        var _this = _super.call(this) || this;
        _this.i = false;
        _this.p = null;
        _this.h = false;
        _this.m = 0;
        _this.l = 0;
        return _this;
    }
    DataGridExpansionIndicatorDescription.prototype.get_type = function () {
        return "DataGridExpansionIndicator";
    };
    Object.defineProperty(DataGridExpansionIndicatorDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridExpansionIndicatorDescription.prototype, "isExpanded", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("IsExpanded");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridExpansionIndicatorDescription.prototype, "iconColor", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("IconColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridExpansionIndicatorDescription.prototype, "isAnimationEnabled", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("IsAnimationEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridExpansionIndicatorDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("PixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataGridExpansionIndicatorDescription.prototype, "actualPixelScalingRatio", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ActualPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    DataGridExpansionIndicatorDescription.$t = markType(DataGridExpansionIndicatorDescription, 'DataGridExpansionIndicatorDescription', Description.$);
    return DataGridExpansionIndicatorDescription;
}(Description));
export { DataGridExpansionIndicatorDescription };
