// igniteui-react-core/esm5/lib/DataLegendSeriesFamily.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Describes type of data columns to use for the data legend.
 */
export var DataLegendSeriesFamily = /*@__PURE__*/ (function (DataLegendSeriesFamily) {
    /**
     * Specifies the Category series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Category"] = 0] = "Category";
    /**
     * Specifies the Range series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Range"] = 1] = "Range";
    /**
     * Specifies the Financial Price series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Financial"] = 2] = "Financial";
    /**
     * Specifies the Financial Indicator series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Indicator"] = 3] = "Indicator";
    /**
     * Specifies the Polar series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Polar"] = 4] = "Polar";
    /**
     * Specifies the Radial series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Radial"] = 5] = "Radial";
    /**
     * Specifies the Scatter series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Scatter"] = 6] = "Scatter";
    /**
     * Specifies the Stacked series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Stacked"] = 7] = "Stacked";
    /**
     * Specifies the Shape series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Shape"] = 8] = "Shape";
    /**
     * Specifies the Geographic series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Geographic"] = 9] = "Geographic";
    /**
     * Specifies the Highlight series
     */
    DataLegendSeriesFamily[DataLegendSeriesFamily["Highlight"] = 10] = "Highlight";
    return DataLegendSeriesFamily;
})({});
/**
 * @hidden
 */
export var DataLegendSeriesFamily_$type = /*@__PURE__*/ markEnum('DataLegendSeriesFamily', 'Category,0|Range,1|Financial,2|Indicator,3|Polar,4|Radial,5|Scatter,6|Stacked,7|Shape,8|Geographic,9|Highlight,10');
