// igniteui-react-core/esm5/lib/DataLegendHeaderTimeMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Describes mode for formatting time in header of data legend
 */
export var DataLegendHeaderTimeMode = /*@__PURE__*/ (function (DataLegendHeaderTimeMode) {
    /**
     * Specifies automatic time format that defaults to no string
     */
    DataLegendHeaderTimeMode[DataLegendHeaderTimeMode["Auto"] = 0] = "Auto";
    /**
     * Specifies time as short string in header row of data legend
     */
    DataLegendHeaderTimeMode[DataLegendHeaderTimeMode["ShortStyle"] = 1] = "ShortStyle";
    /**
     * Specifies time as long string in header row of data legend
     */
    DataLegendHeaderTimeMode[DataLegendHeaderTimeMode["LongStyle"] = 2] = "LongStyle";
    /**
     * Specifies time as medium string in header row of data legend
     */
    DataLegendHeaderTimeMode[DataLegendHeaderTimeMode["MediumStyle"] = 3] = "MediumStyle";
    /**
     * Specifies time as full string in header row of data legend
     */
    DataLegendHeaderTimeMode[DataLegendHeaderTimeMode["FullStyle"] = 4] = "FullStyle";
    /**
     * Specifies no time in header row of data legend
     */
    DataLegendHeaderTimeMode[DataLegendHeaderTimeMode["None"] = 5] = "None";
    return DataLegendHeaderTimeMode;
})({});
/**
 * @hidden
 */
export var DataLegendHeaderTimeMode_$type = /*@__PURE__*/ markEnum('DataLegendHeaderTimeMode', 'Auto,0|ShortStyle,1|LongStyle,2|MediumStyle,3|FullStyle,4|None,5');
