// igniteui-react-core/esm5/lib/DataLegendSummaryType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Describes type of summarization to use for the data legend.
 */
export var DataLegendSummaryType = /*@__PURE__*/ (function (DataLegendSummaryType) {
    /**
     * Specifies automatic summarization
     */
    DataLegendSummaryType[DataLegendSummaryType["Auto"] = 0] = "Auto";
    /**
     * Specifies total of values in summary row
     */
    DataLegendSummaryType[DataLegendSummaryType["Total"] = 1] = "Total";
    /**
     * Specifies average of values in summary row
     */
    DataLegendSummaryType[DataLegendSummaryType["Average"] = 2] = "Average";
    /**
     * Specifies minimum of values in summary row
     */
    DataLegendSummaryType[DataLegendSummaryType["Min"] = 3] = "Min";
    /**
     * Specifies maximum of values in summary row
     */
    DataLegendSummaryType[DataLegendSummaryType["Max"] = 4] = "Max";
    /**
     * Specifies no summarization
     */
    DataLegendSummaryType[DataLegendSummaryType["None"] = 5] = "None";
    return DataLegendSummaryType;
})({});
/**
 * @hidden
 */
export var DataLegendSummaryType_$type = /*@__PURE__*/ markEnum('DataLegendSummaryType', 'Auto,0|Total,1|Average,2|Min,3|Max,4|None,5');
