// igniteui-react-core/esm5/lib/DataVisualizationLocalePl.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocalePl = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocalePl() {
        this["AssemblyNotIncluded"] = "Nie można załadować pliku {0}.dll. Upewnij się, że ten zestaw jest uwzględniony w pliku XAP, który jest wdrażany w systemie Windows Phone, na przykład przez dodanie do niego odwołania z dowolnego projektu.";
        this["BreakEvenTitle_BreakEven"] = "Na czysto";
        this["BreakEvenTitle_FixedCost"] = "Koszt stały";
        this["BreakEvenTitle_LossArea"] = "Obszar strat";
        this["BreakEvenTitle_MarginalProfit"] = "Zysk krańcowy";
        this["BreakEvenTitle_ProfitArea"] = "Obszar zysków";
        this["BreakEvenTitle_SafetyMargin"] = "Margines bezpieczeństwa";
        this["BreakEvenTitle_SalesRevenue"] = "Przychody ze sprzedaży";
        this["BreakEvenTitle_TotalCost"] = "Koszt całkowity";
        this["BreakEvenTitle_VariableCost"] = "Koszt zmienny";
        this["BubbleSeries_Radius"] = "Promień";
        this["DataChart_InteractivityNotLoaded"] = "Uzyskano dostęp do funkcji interaktywności, ale moduł interakcji z wykresem nie został załadowany.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Użyto funkcji skracania liczb, ale moduł skracania liczb nie został załadowany.";
        this["DataChart_VisualDataNotLoaded"] = "Uzyskano dostęp do funkcji VisualData, ale moduł interakcji z wykresem nie został załadowany.";
        this["DataProviderNotInitialized"] = "Dostawca danych został użyty bez wcześniejszego zainicjowania. Dostawca danych musi zostać zainicjowany przed użyciem.";
        this["DataSource_Summary_Avg"] = "Śr.";
        this["DataSource_Summary_Count"] = "Liczba";
        this["DataSource_Summary_Max"] = "Maks";
        this["DataSource_Summary_Min"] = "Min.";
        this["DataSource_Summary_Sum"] = "Suma";
        this["Default_Series_Title"] = "Tytuły serii danych";
        this["FinancialChart_IndicatorMenu_Header"] = "Wskaźniki";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Oddzielne wskaźniki";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Nakładki";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Linie trendów";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Woluminy";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Oscylator ilości bezwzględnej";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Akumulacja/Dystrybucja";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Średni wskaźnik ruchu kierunkowego";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Średni prawdziwy zakres zmiany";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Szerokość pasm Bollingera";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Oscylator Chaikina";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Zmienność Chaikina";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Wskaźnik kanału towaru";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Beztrendowy oscylator cenowy (wskaźnik DPO)";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Wskaźnik swobody ruchu";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Szybki oscylator stochastyczny";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Wskaźnik siły";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Pełny oscylator stochastyczny";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Wskaźnik MFI";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Wskaźnik masy";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Mediana ceny";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Wskaźnik przepływów pieniężnych";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Wskaźnik wolumenu ujemnego";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "Wskaźnik OBV";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Oscylator ceny procentowej (PPO)";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Oscylator wolumenu procentowego (PVO)";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Wskaźnik wolumenu dodatniego";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Trend wolumenu cen (PVT)";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Wolny oscylator stochastyczny";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Odchylenie standardowe";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Cena typowa";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Oscylator UO";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Ważone zamknięcie";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "%R Williamsa";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Wstęgi Bollingera";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Cena kanału";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Dopasowanie sześcienne";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Średnia skumulowana";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Średnia wykładnicza";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Dopasowanie wykładnicze";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Dopasowanie liniowe";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Dopasowanie logarytmiczne";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Średnia zmodyfikowana";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "Dopasowanie PowerLaw";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Dopasowanie kwadratowe";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Dopasowanie czwartego stopnia";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Dopasowanie piątego stopnia";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Średnia zwykła";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Średnia ważona";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Obszar";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Kolumna";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Linia";
        this["FinancialChart_RangeSelector_All"] = "Wszystko";
        this["FinancialChart_RangeSelector_From"] = "Od";
        this["FinancialChart_RangeSelector_OneMonth"] = "1M";
        this["FinancialChart_RangeSelector_OneYear"] = "1Y";
        this["FinancialChart_RangeSelector_SixMonths"] = "6M";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3M";
        this["FinancialChart_RangeSelector_To"] = "Do";
        this["FinancialChart_RangeSelector_YearToDate"] = "YTD";
        this["FinancialSeries_Close"] = "Cena zamknięcia";
        this["FinancialSeries_High"] = "Szczyt";
        this["FinancialSeries_Low"] = "Dołek";
        this["FinancialSeries_Open"] = "Cena otwarcia";
        this["FinancialSeries_Volume"] = "Wolumen";
        this["NoEncodingsLoaded"] = "Nie znaleziono żadnych kodowań! Załaduj zestaw Infragistics.Encoding lub zapoznaj się z dokumentacją dotyczącą tworzenia niestandardowych kodowań.";
        this["NotSupportedEncoding"] = "Kodowanie odpowiadające {0} {1} nie jest obsługiwane.";
        this["Object_Sealed"] = "Po zapieczętowaniu obiektu modyfikacje są niedozwolone.";
        this["OPD_DefaultInteraction"] = "Wybierz domyślną interakcję przeciągania myszą";
        this["OPD_DragPan"] = "Przesuwaj powierzchnię";
        this["OPD_DragZoom"] = "Powiększ obszar wyboru";
        this["OPD_ScaleToFit"] = "Dopasuj do ekranu";
        this["OPD_ScaleToFit_SeriesViewer"] = "Resetuj powiększenie (100%)";
        this["OPD_ZoomIn"] = "Powiększ";
        this["OPD_ZoomOut"] = "Pomniejsz";
        this["OPD_ZoomTo100"] = "Powiększenie do 100%";
        this["PieChart_Others"] = "Inne";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Modyfikacje źródła danych oparte na zakresie nie są obecnie obsługiwane";
        this["ScatterSeries_Value"] = "Wartość";
        this["TRIAL_VERSION"] = "Zakup NIELICENCJONOWANEJ WERSJI PRÓBNEJ na Infragistics.com";
    }
    return DataVisualizationLocalePl;
}());
export { DataVisualizationLocalePl };
