// igniteui-react-core/esm5/lib/DataPieChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DataPieBaseChartDescription } from "./DataPieBaseChartDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var DataPieChartDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataPieChartDescription, _super);
    function DataPieChartDescription() {
        var _this = _super.call(this) || this;
        _this.yd = 0;
        _this.ye = 0;
        _this.yl = null;
        _this.yo = null;
        _this.yi = null;
        _this.yb = false;
        _this.ym = null;
        _this.yn = null;
        _this.yk = null;
        _this.yp = null;
        _this.yj = null;
        _this.yh = null;
        return _this;
    }
    DataPieChartDescription.prototype.get_type = function () {
        return "DataPieChart";
    };
    Object.defineProperty(DataPieChartDescription.prototype, "transitionInDuration", {
        get: function () {
            return this.yd;
        },
        set: function (a) {
            this.yd = a;
            this.g("TransitionInDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "transitionOutDuration", {
        get: function () {
            return this.ye;
        },
        set: function (a) {
            this.ye = a;
            this.g("TransitionOutDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "transitionInEasingFunctionRef", {
        get: function () {
            return this.yl;
        },
        set: function (a) {
            this.yl = a;
            this.g("TransitionInEasingFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "transitionOutEasingFunctionRef", {
        get: function () {
            return this.yo;
        },
        set: function (a) {
            this.yo = a;
            this.g("TransitionOutEasingFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "chartType", {
        get: function () {
            return this.yi;
        },
        set: function (a) {
            this.yi = a;
            this.g("ChartType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "isTransitionInEnabled", {
        get: function () {
            return this.yb;
        },
        set: function (a) {
            this.yb = a;
            this.g("IsTransitionInEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "transitionInMode", {
        get: function () {
            return this.ym;
        },
        set: function (a) {
            this.ym = a;
            this.g("TransitionInMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "transitionInSpeedType", {
        get: function () {
            return this.yn;
        },
        set: function (a) {
            this.yn = a;
            this.g("TransitionInSpeedType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "tooltipTemplateRef", {
        get: function () {
            return this.yk;
        },
        set: function (a) {
            this.yk = a;
            this.g("TooltipTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "width", {
        get: function () {
            return this.yp;
        },
        set: function (a) {
            this.yp = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "height", {
        get: function () {
            return this.yj;
        },
        set: function (a) {
            this.yj = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "background", {
        get: function () {
            return this.yh;
        },
        set: function (a) {
            this.yh = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    DataPieChartDescription.$t = markType(DataPieChartDescription, 'DataPieChartDescription', DataPieBaseChartDescription.$);
    return DataPieChartDescription;
}(DataPieBaseChartDescription));
export { DataPieChartDescription };
