// igniteui-react-core/esm5/lib/DataSourceSpecialRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataSourceSpecialRowDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataSourceSpecialRowDescription, _super);
    function DataSourceSpecialRowDescription() {
        var _this = _super.call(this) || this;
        _this.p = null;
        _this.m = null;
        _this.k = 0;
        _this.h = null;
        return _this;
    }
    DataSourceSpecialRowDescription.prototype.get_type = function () {
        return "DataSourceSpecialRow";
    };
    Object.defineProperty(DataSourceSpecialRowDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataSourceSpecialRowDescription.prototype, "rowType", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("RowType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataSourceSpecialRowDescription.prototype, "targetRow", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("TargetRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataSourceSpecialRowDescription.prototype, "level", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Level");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataSourceSpecialRowDescription.prototype, "summaryResults", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("SummaryResults");
        },
        enumerable: false,
        configurable: true
    });
    DataSourceSpecialRowDescription.$t = markType(DataSourceSpecialRowDescription, 'DataSourceSpecialRowDescription', Description.$);
    DataSourceSpecialRowDescription.__marshalByValue = true;
    DataSourceSpecialRowDescription.__marshalByValueAlias = "DataSourceSpecialRow";
    return DataSourceSpecialRowDescription;
}(Description));
export { DataSourceSpecialRowDescription };
