// igniteui-react-core/esm5/lib/DataVisualizationLocaleZhHant.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
var DataVisualizationLocaleZhHant = /** @class */ /*@__PURE__*/ (function () {
    function DataVisualizationLocaleZhHant() {
        this["AssemblyNotIncluded"] = "無法載入 {0}.dll。您需要確保此程序集包含在已部署至 Windows Phone 的 XAP 中，例如，從您的任何專案中添加對該程序集的引用。";
        this["BreakEvenTitle_BreakEven"] = "收支平衡";
        this["BreakEvenTitle_FixedCost"] = "固定成本";
        this["BreakEvenTitle_LossArea"] = "損失面積";
        this["BreakEvenTitle_MarginalProfit"] = "邊際利潤";
        this["BreakEvenTitle_ProfitArea"] = "利潤範圍";
        this["BreakEvenTitle_SafetyMargin"] = "安全寬限期";
        this["BreakEvenTitle_SalesRevenue"] = "銷售收入";
        this["BreakEvenTitle_TotalCost"] = "總費用";
        this["BreakEvenTitle_VariableCost"] = "變動成本";
        this["BubbleSeries_Radius"] = "半徑";
        this["DataChart_InteractivityNotLoaded"] = "已存取互動功能，但尚未載入圖表互動模組。";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "已使用數字縮寫功能，但尚未加載數字縮寫模組。";
        this["DataChart_VisualDataNotLoaded"] = "已存取 VisualData 功能，但尚未載入圖表互動模組。";
        this["DataProviderNotInitialized"] = "未先初始化就使用了資料提供者。資料提供者必須在使用前初始化。";
        this["DataSource_Summary_Avg"] = "平均";
        this["DataSource_Summary_Count"] = "計數";
        this["DataSource_Summary_Max"] = "最大";
        this["DataSource_Summary_Min"] = "最小";
        this["DataSource_Summary_Sum"] = "總和";
        this["Default_Series_Title"] = "數列標題";
        this["FinancialChart_IndicatorMenu_Header"] = "指標";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "獨立指標";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "疊加";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "趨勢線";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "成交量";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "絕對成交量震盪指標";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "累積分佈";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "平均方向指數";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "平均真實範圍";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "布林帶寬度";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "柴金震盪指標";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "柴金波動率";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "商品通道指數";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "去趨勢價格震盪指標";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "輕鬆移動";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "快速隨機震盪指標";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "力指數";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "全隨機震盪指標";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "市場促進指數";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "質量指數";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "中位數價格";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "資金流量指數";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "負成交量指數";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "平衡成交量指標";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "百分比價格震盪指標";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "百分比成交量震盪指標";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "正成交量指數";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "價格量趨勢";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "變動率 (Rate of Change)";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "相對強弱指標 (Relative Strength Index)";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "慢速隨機震盪指標";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "標準差";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "随機相對强弱指數";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "典型價格";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "終極震盪指標";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "加權關閉";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "威廉指數";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "布林帶";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "價格頻道";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "立方體調整";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "累計平均值";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "指數平均值";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "指數調整";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "線性調整";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "对数調整";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "修正平均值";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "幂律擬合";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "二次方調整";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "四次方調整";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "五次方調整";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "簡單平均值";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "加權平均值";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "區域";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "列";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "線";
        this["FinancialChart_RangeSelector_All"] = "全部";
        this["FinancialChart_RangeSelector_From"] = "自";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 個月";
        this["FinancialChart_RangeSelector_OneYear"] = "1 年";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 個月";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 個月";
        this["FinancialChart_RangeSelector_To"] = "至";
        this["FinancialChart_RangeSelector_YearToDate"] = "年初至今";
        this["FinancialSeries_Close"] = "關閉";
        this["FinancialSeries_High"] = "高";
        this["FinancialSeries_Low"] = "低";
        this["FinancialSeries_Open"] = "開啟";
        this["FinancialSeries_Volume"] = "成交量";
        this["NoEncodingsLoaded"] = "未找到任何編碼！請加載 Infragistics.Encoding 程序集或查看文檔，了解如何創建自定義編碼。";
        this["NotSupportedEncoding"] = "不支援 {0} {1} 下的對應編碼。";
        this["Object_Sealed"] = "密封物體後，不允許修改。";
        this["OPD_DefaultInteraction"] = "選擇默認的鼠標拖曳互動";
        this["OPD_DragPan"] = "平移表面";
        this["OPD_DragZoom"] = "縮放選取範圍";
        this["OPD_ScaleToFit"] = "全螢幕";
        this["OPD_ScaleToFit_SeriesViewer"] = "重設縮放 (100%)";
        this["OPD_ZoomIn"] = "放大";
        this["OPD_ZoomOut"] = "縮小";
        this["OPD_ZoomTo100"] = "縮放成 100%";
        this["PieChart_Others"] = "其他";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "當前不支持基於範圍的數據源修改";
        this["ScatterSeries_Value"] = "數值";
        this["TRIAL_VERSION"] = "未授權的試用版，在 Infragistics.com 購買";
    }
    return DataVisualizationLocaleZhHant;
}());
export { DataVisualizationLocaleZhHant };
