// igniteui-react-core/esm5/lib/DataSourceAggregatedResultDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataSourceAggregatedResultDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataSourceAggregatedResultDescription, _super);
    function DataSourceAggregatedResultDescription() {
        var _this = _super.call(this) || this;
        _this.l = null;
        _this.m = null;
        _this.h = null;
        _this.n = null;
        return _this;
    }
    DataSourceAggregatedResultDescription.prototype.get_type = function () {
        return "DataSourceAggregatedResult";
    };
    Object.defineProperty(DataSourceAggregatedResultDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataSourceAggregatedResultDescription.prototype, "itemRef", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataSourceAggregatedResultDescription.prototype, "transactionType", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("TransactionType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataSourceAggregatedResultDescription.prototype, "keys", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Keys");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataSourceAggregatedResultDescription.prototype, "valuesRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ValuesRef");
        },
        enumerable: false,
        configurable: true
    });
    DataSourceAggregatedResultDescription.$t = markType(DataSourceAggregatedResultDescription, 'DataSourceAggregatedResultDescription', Description.$);
    DataSourceAggregatedResultDescription.__marshalByValue = true;
    DataSourceAggregatedResultDescription.__marshalByValueAlias = "DataSourceAggregatedResult";
    return DataSourceAggregatedResultDescription;
}(Description));
export { DataSourceAggregatedResultDescription };
