// igniteui-react-core/esm5/lib/DateTimeColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DataGridColumnDescription } from "./DataGridColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var DateTimeColumnDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DateTimeColumnDescription, _super);
    function DateTimeColumnDescription() {
        var _this = _super.call(this) || this;
        _this.f9 = null;
        _this.gg = null;
        _this.gb = null;
        _this.f1 = null;
        _this.gf = null;
        _this.f6 = false;
        _this.gd = null;
        _this.ga = null;
        _this.f3 = null;
        _this.gc = null;
        _this.ge = null;
        return _this;
    }
    DateTimeColumnDescription.prototype.get_type = function () {
        return "DateTimeColumn";
    };
    Object.defineProperty(DateTimeColumnDescription.prototype, "dateTimeFormat", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("DateTimeFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeColumnDescription.prototype, "formatString", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("FormatString");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeColumnDescription.prototype, "editorFormatString", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("EditorFormatString");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeColumnDescription.prototype, "formatSpecifiers", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("FormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeColumnDescription.prototype, "formatOverrideRef", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("FormatOverrideRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeColumnDescription.prototype, "showTodayButton", {
        get: function () {
            return this.f6;
        },
        set: function (a) {
            this.f6 = a;
            this.g("ShowTodayButton");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeColumnDescription.prototype, "editorType", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("EditorType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeColumnDescription.prototype, "editorDataSourceRef", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("EditorDataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeColumnDescription.prototype, "editorDataSource", {
        get: function () {
            return this.f3;
        },
        set: function (a) {
            this.f3 = a;
            this.g("EditorDataSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeColumnDescription.prototype, "editorTextField", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("EditorTextField");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeColumnDescription.prototype, "editorValueField", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("EditorValueField");
        },
        enumerable: false,
        configurable: true
    });
    DateTimeColumnDescription.$t = markType(DateTimeColumnDescription, 'DateTimeColumnDescription', DataGridColumnDescription.$);
    DateTimeColumnDescription.__canMarshalByValue1 = true;
    DateTimeColumnDescription.__marshalByValueAlias1 = "DateTimeColumn";
    return DateTimeColumnDescription;
}(DataGridColumnDescription));
export { DateTimeColumnDescription };
