// igniteui-react-core/esm5/lib/DateTimeFormatSpecifierDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { FormatSpecifierDescription } from "./FormatSpecifierDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var DateTimeFormatSpecifierDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DateTimeFormatSpecifierDescription, _super);
    function DateTimeFormatSpecifierDescription() {
        var _this = _super.call(this) || this;
        _this.x = null;
        _this.p = null;
        _this.ad = null;
        _this.l = 0;
        _this.o = null;
        _this.r = null;
        _this.ab = null;
        _this.y = null;
        _this.ae = null;
        _this.v = null;
        _this.w = null;
        _this.t = null;
        _this.ag = null;
        _this.s = null;
        _this.ah = null;
        _this.aa = null;
        _this.q = null;
        _this.u = null;
        _this.z = null;
        _this.ac = null;
        _this.af = null;
        return _this;
    }
    DateTimeFormatSpecifierDescription.prototype.get_type = function () {
        return "DateTimeFormatSpecifier";
    };
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "locale", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("Locale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "dateStyle", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("DateStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "timeStyle", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("TimeStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "fractionalSecondDigits", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("FractionalSecondDigits");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "calendar", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Calendar");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "dayPeriod", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("DayPeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "numberingSystem", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("NumberingSystem");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "localeMatcher", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("LocaleMatcher");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "timeZone", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("TimeZone");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "hour12", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("Hour12");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "hourCycle", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("HourCycle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "formatMatcher", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("FormatMatcher");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "weekDay", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("WeekDay");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "era", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Era");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "year", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("Year");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "month", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("Month");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "day", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Day");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "hour", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Hour");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "minute", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("Minute");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "second", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("Second");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DateTimeFormatSpecifierDescription.prototype, "timeZoneName", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("TimeZoneName");
        },
        enumerable: false,
        configurable: true
    });
    DateTimeFormatSpecifierDescription.$t = markType(DateTimeFormatSpecifierDescription, 'DateTimeFormatSpecifierDescription', FormatSpecifierDescription.$);
    DateTimeFormatSpecifierDescription.__marshalByValue1 = true;
    DateTimeFormatSpecifierDescription.__marshalByValueAlias1 = "DateTimeFormatSpecifier";
    return DateTimeFormatSpecifierDescription;
}(FormatSpecifierDescription));
export { DateTimeFormatSpecifierDescription };
