// igniteui-react-core/esm5/lib/DatePartDeltasDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DatePartDeltasDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DatePartDeltasDescription, _super);
    function DatePartDeltasDescription() {
        var _this = _super.call(this) || this;
        _this.h = 0;
        _this.k = 0;
        _this.m = 0;
        _this.i = 0;
        _this.j = 0;
        _this.l = 0;
        return _this;
    }
    DatePartDeltasDescription.prototype.get_type = function () {
        return "DatePartDeltas";
    };
    Object.defineProperty(DatePartDeltasDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePartDeltasDescription.prototype, "date", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Date");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePartDeltasDescription.prototype, "month", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Month");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePartDeltasDescription.prototype, "year", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Year");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePartDeltasDescription.prototype, "hours", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Hours");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePartDeltasDescription.prototype, "minutes", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Minutes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DatePartDeltasDescription.prototype, "seconds", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Seconds");
        },
        enumerable: false,
        configurable: true
    });
    DatePartDeltasDescription.$t = markType(DatePartDeltasDescription, 'DatePartDeltasDescription', Description.$);
    return DatePartDeltasDescription;
}(Description));
export { DatePartDeltasDescription };
