// igniteui-react-core/esm5/lib/Decoder.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, markType } from "./type";
import { NotImplementedException } from "./NotImplementedException";
/**
 * @hidden
 */
var Decoder = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Decoder, _super);
    function Decoder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Decoder.prototype.e = function (a, b, c, d, e, f, g, h, i, j) {
        throw new NotImplementedException(0);
        return {
            p7: h,
            p8: i,
            p9: j
        };
    };
    Decoder.prototype.b = function (a, b, c, d) {
        throw new NotImplementedException(0);
    };
    Decoder.prototype.d = function (a, b, c, d, e, f) {
        throw new NotImplementedException(0);
    };
    Decoder.prototype.f = function () {
        throw new NotImplementedException(0);
    };
    Decoder.$t = markType(Decoder, 'Decoder');
    return Decoder;
}(Base));
export { Decoder };
