// igniteui-react-core/esm5/lib/DivBridge.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { NativeUIElementBridge } from "./NativeUIElementBridge";
import { INativeUIContentBridge_$type } from "./INativeUIContentBridge";
import { markType } from "./type";
/**
 * @hidden
 */
var DivBridge = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DivBridge, _super);
    function DivBridge() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.s = null;
        return _this;
    }
    DivBridge.prototype.getContent = function (a) {
        return this.s;
    };
    DivBridge.prototype.getValue = function (a, b) {
        return null;
    };
    DivBridge.prototype.setContent = function (a, b) {
        if (this.s == b) {
            return;
        }
        this.s = b;
        if (b == null) {
            for (var c = a.getChildCount() - 1; c >= 0; c--) {
                var d = a.getChildAt(c);
                d.remove();
            }
        }
        else {
            for (var e = a.getChildCount() - 1; e >= 0; e--) {
                var f = a.getChildAt(e);
                f.remove();
            }
            a.append(b.an);
        }
    };
    DivBridge.prototype.setValue = function (a, b, c) {
    };
    DivBridge.$t = markType(DivBridge, 'DivBridge', NativeUIElementBridge.$, [INativeUIContentBridge_$type]);
    return DivBridge;
}(NativeUIElementBridge));
export { DivBridge };
