// igniteui-react-core/esm5/lib/DomainChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DomainChartDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DomainChartDescription, _super);
    function DomainChartDescription() {
        var _this = _super.call(this) || this;
        _this.fg = 0;
        _this.ft = 0;
        _this.fu = 0;
        _this.fv = 0;
        _this.fs = 0;
        _this.fo = 0;
        _this.fq = 0;
        _this.fp = 0;
        _this.fn = 0;
        _this.nl = null;
        _this.no = null;
        _this.fd = 0;
        _this.fw = 0;
        _this.fm = 0;
        _this.dc = 0;
        _this.jk = 0;
        _this.nr = null;
        _this.jh = 0;
        _this.jj = 0;
        _this.jg = 0;
        _this.nm = null;
        _this.np = null;
        _this.kk = null;
        _this.d9 = 0;
        _this.ea = 0;
        _this.dj = 0;
        _this.dk = 0;
        _this.ki = null;
        _this.kj = null;
        _this.b3 = false;
        _this.p = null;
        _this.m = null;
        _this.o = null;
        _this.l = null;
        _this.lx = null;
        _this.jd = 0;
        _this.je = 0;
        _this.lz = null;
        _this.ly = null;
        _this.b4 = false;
        _this.l0 = null;
        _this.q = null;
        _this.ex = 0;
        _this.ey = 0;
        _this.ez = 0;
        _this.e0 = 0;
        _this.b5 = false;
        _this.l5 = null;
        _this.b6 = false;
        _this.e2 = 0;
        _this.e3 = 0;
        _this.e4 = 0;
        _this.e5 = 0;
        _this.l1 = null;
        _this.l2 = null;
        _this.e1 = 0;
        _this.l3 = null;
        _this.l4 = null;
        _this.kr = null;
        _this.n = null;
        _this.kp = null;
        _this.kq = null;
        _this.ks = null;
        _this.kt = null;
        _this.ku = null;
        _this.d0 = 0;
        _this.d1 = 0;
        _this.d2 = 0;
        _this.d3 = 0;
        _this.dv = 0;
        _this.dw = 0;
        _this.dx = 0;
        _this.dy = 0;
        _this.b2 = false;
        _this.kv = null;
        _this.dz = 0;
        _this.kw = null;
        _this.kx = null;
        _this.kl = null;
        _this.dr = 0;
        _this.ds = 0;
        _this.dt = 0;
        _this.du = 0;
        _this.dl = 0;
        _this.dm = 0;
        _this.dn = 0;
        _this.dp = 0;
        _this.b1 = false;
        _this.km = null;
        _this.dq = 0;
        _this.kn = null;
        _this.ko = null;
        _this.k9 = null;
        _this.la = null;
        _this.eg = 0;
        _this.lb = null;
        _this.lc = null;
        _this.ld = null;
        _this.k8 = null;
        _this.eh = 0;
        _this.ei = 0;
        _this.ej = 0;
        _this.ek = 0;
        _this.ec = 0;
        _this.ed = 0;
        _this.ee = 0;
        _this.ef = 0;
        _this.lj = null;
        _this.lk = null;
        _this.em = 0;
        _this.ll = null;
        _this.lm = null;
        _this.k3 = null;
        _this.k4 = null;
        _this.k5 = null;
        _this.eb = 0;
        _this.k6 = null;
        _this.k7 = null;
        _this.le = null;
        _this.lf = null;
        _this.lg = null;
        _this.el = 0;
        _this.lh = null;
        _this.li = null;
        _this.df = 0;
        _this.dg = 0;
        _this.dh = 0;
        _this.di = 0;
        _this.kh = null;
        _this.lr = null;
        _this.ls = null;
        _this.lt = null;
        _this.et = 0;
        _this.eu = 0;
        _this.ev = 0;
        _this.ew = 0;
        _this.lu = null;
        _this.es = 0;
        _this.lv = null;
        _this.lw = null;
        _this.eo = 0;
        _this.ep = 0;
        _this.eq = 0;
        _this.er = 0;
        _this.ln = null;
        _this.lo = null;
        _this.en = 0;
        _this.lp = null;
        _this.lq = null;
        _this.ky = null;
        _this.kz = null;
        _this.d5 = 0;
        _this.d6 = 0;
        _this.d7 = 0;
        _this.d8 = 0;
        _this.k0 = null;
        _this.d4 = 0;
        _this.k1 = null;
        _this.k2 = null;
        _this.aj = null;
        _this.v = null;
        _this.u = null;
        _this.s = null;
        _this.w = null;
        _this.ak = null;
        _this.m8 = null;
        _this.mb = null;
        _this.m7 = null;
        _this.l9 = null;
        _this.m6 = null;
        _this.ai = null;
        _this.t = null;
        _this.ac = null;
        _this.aa = null;
        _this.z = null;
        _this.y = null;
        _this.ab = null;
        _this.ad = null;
        _this.mp = null;
        _this.mc = null;
        _this.mn = null;
        _this.mm = null;
        _this.mo = null;
        _this.mq = null;
        _this.kg = null;
        _this.bl = null;
        _this.md = null;
        _this.bm = null;
        _this.x = null;
        _this.r = null;
        _this.j = null;
        _this.ah = null;
        _this.h = null;
        _this.i = null;
        _this.mv = null;
        _this.b9 = false;
        _this.cb = false;
        _this.ce = false;
        _this.cg = false;
        _this.mf = null;
        _this.ca = false;
        _this.me = null;
        _this.mu = null;
        _this.bp = null;
        _this.j7 = null;
        _this.nj = null;
        _this.nn = null;
        _this.nk = null;
        _this.nt = null;
        _this.fl = 0;
        _this.fr = 0;
        _this.mz = null;
        _this.my = null;
        _this.mx = null;
        _this.fe = 0;
        _this.ff = 0;
        _this.ag = null;
        _this.mw = null;
        _this.ae = null;
        _this.af = null;
        _this.ji = 0;
        _this.db = 0;
        _this.bs = false;
        _this.al = null;
        _this.ns = null;
        _this.fx = 0;
        _this.br = false;
        _this.fi = 0;
        _this.fk = 0;
        _this.fj = 0;
        _this.fh = 0;
        _this.j8 = null;
        _this.nb = null;
        _this.nc = null;
        _this.mh = null;
        _this.mg = null;
        _this.mr = null;
        _this.mt = null;
        _this.ms = null;
        _this.jl = 0;
        _this.nq = null;
        _this.kd = null;
        _this.b0 = false;
        _this.kf = null;
        _this.de = 0;
        _this.ke = null;
        _this.bz = false;
        _this.j9 = null;
        _this.kb = null;
        _this.ka = null;
        _this.kc = null;
        _this.jb = 0;
        _this.jc = 0;
        _this.cd = false;
        _this.cf = false;
        _this.b7 = false;
        _this.l7 = null;
        _this.l8 = null;
        _this.jf = 0;
        _this.bt = false;
        _this.by = false;
        _this.bu = false;
        _this.ch = false;
        _this.jy = null;
        _this.bw = false;
        _this.bx = false;
        _this.bv = false;
        _this.dd = 0;
        _this.j0 = null;
        _this.ja = 0;
        _this.j2 = null;
        _this.k = null;
        _this.j5 = null;
        _this.j6 = null;
        _this.j3 = null;
        _this.j1 = null;
        _this.am = null;
        _this.an = null;
        _this.fy = 0;
        _this.cc = false;
        _this.mj = null;
        _this.nv = null;
        _this.ml = null;
        _this.nx = null;
        _this.mi = null;
        _this.mk = null;
        _this.fa = 0;
        _this.e9 = 0;
        _this.e6 = 0;
        _this.e7 = 0;
        _this.e8 = 0;
        _this.fc = 0;
        _this.fb = 0;
        _this.b8 = false;
        _this.nu = null;
        _this.nw = null;
        _this.f2 = 0;
        _this.f1 = 0;
        _this.fz = 0;
        _this.f5 = 0;
        _this.f0 = 0;
        _this.f4 = 0;
        _this.f3 = 0;
        _this.ci = false;
        _this.f7 = 0;
        _this.f6 = 0;
        _this.f9 = 0;
        _this.f8 = 0;
        _this.m9 = null;
        _this.ni = null;
        _this.ne = null;
        _this.nf = null;
        _this.ng = null;
        _this.nd = null;
        _this.nh = null;
        _this.na = null;
        _this.m1 = null;
        _this.m2 = null;
        _this.m3 = null;
        _this.m0 = null;
        _this.m4 = null;
        _this.j4 = null;
        _this.jz = null;
        _this.m5 = null;
        _this.ma = null;
        _this.l6 = null;
        return _this;
    }
    DomainChartDescription.prototype.get_type = function () {
        return "DomainChart";
    };
    Object.defineProperty(DomainChartDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.fg;
        },
        set: function (a) {
            this.fg = a;
            this.g("PixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "titleLeftMargin", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("TitleLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "titleRightMargin", {
        get: function () {
            return this.fu;
        },
        set: function (a) {
            this.fu = a;
            this.g("TitleRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "titleTopMargin", {
        get: function () {
            return this.fv;
        },
        set: function (a) {
            this.fv = a;
            this.g("TitleTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "titleBottomMargin", {
        get: function () {
            return this.fs;
        },
        set: function (a) {
            this.fs = a;
            this.g("TitleBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "subtitleLeftMargin", {
        get: function () {
            return this.fo;
        },
        set: function (a) {
            this.fo = a;
            this.g("SubtitleLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "subtitleTopMargin", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("SubtitleTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "subtitleRightMargin", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("SubtitleRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "subtitleBottomMargin", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("SubtitleBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "subtitleTextColor", {
        get: function () {
            return this.nl;
        },
        set: function (a) {
            this.nl = a;
            this.g("SubtitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "titleTextColor", {
        get: function () {
            return this.no;
        },
        set: function (a) {
            this.no = a;
            this.g("TitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "leftMargin", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("LeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "topMargin", {
        get: function () {
            return this.fw;
        },
        set: function (a) {
            this.fw = a;
            this.g("TopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "rightMargin", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("RightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "bottomMargin", {
        get: function () {
            return this.dc;
        },
        set: function (a) {
            this.dc = a;
            this.g("BottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "transitionDuration", {
        get: function () {
            return this.jk;
        },
        set: function (a) {
            this.jk = a;
            this.g("TransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "transitionEasingFunctionRef", {
        get: function () {
            return this.nr;
        },
        set: function (a) {
            this.nr = a;
            this.g("TransitionEasingFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "highlightingTransitionDuration", {
        get: function () {
            return this.jh;
        },
        set: function (a) {
            this.jh = a;
            this.g("HighlightingTransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "selectionTransitionDuration", {
        get: function () {
            return this.jj;
        },
        set: function (a) {
            this.jj = a;
            this.g("SelectionTransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "focusTransitionDuration", {
        get: function () {
            return this.jg;
        },
        set: function (a) {
            this.jg = a;
            this.g("FocusTransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "subtitleTextStyle", {
        get: function () {
            return this.nm;
        },
        set: function (a) {
            this.nm = a;
            this.g("SubtitleTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "titleTextStyle", {
        get: function () {
            return this.np;
        },
        set: function (a) {
            this.np = a;
            this.g("TitleTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupingMode", {
        get: function () {
            return this.kk;
        },
        set: function (a) {
            this.kk = a;
            this.g("DataToolTipGroupingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipPositionOffsetX", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("DataToolTipPositionOffsetX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipPositionOffsetY", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.g("DataToolTipPositionOffsetY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipDefaultPositionOffsetX", {
        get: function () {
            return this.dj;
        },
        set: function (a) {
            this.dj = a;
            this.g("DataToolTipDefaultPositionOffsetX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipDefaultPositionOffsetY", {
        get: function () {
            return this.dk;
        },
        set: function (a) {
            this.dk = a;
            this.g("DataToolTipDefaultPositionOffsetY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupedPositionModeX", {
        get: function () {
            return this.ki;
        },
        set: function (a) {
            this.ki = a;
            this.g("DataToolTipGroupedPositionModeX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupedPositionModeY", {
        get: function () {
            return this.kj;
        },
        set: function (a) {
            this.kj = a;
            this.g("DataToolTipGroupedPositionModeY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipShouldUpdateWhenSeriesDataChanges", {
        get: function () {
            return this.b3;
        },
        set: function (a) {
            this.b3 = a;
            this.g("DataToolTipShouldUpdateWhenSeriesDataChanges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipIncludedSeries", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("DataToolTipIncludedSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipExcludedSeries", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("DataToolTipExcludedSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipIncludedColumns", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("DataToolTipIncludedColumns");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipExcludedColumns", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("DataToolTipExcludedColumns");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueFormatAbbreviation", {
        get: function () {
            return this.lx;
        },
        set: function (a) {
            this.lx = a;
            this.g("DataToolTipValueFormatAbbreviation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueFormatMaxFractions", {
        get: function () {
            return this.jd;
        },
        set: function (a) {
            this.jd = a;
            this.g("DataToolTipValueFormatMaxFractions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueFormatMinFractions", {
        get: function () {
            return this.je;
        },
        set: function (a) {
            this.je = a;
            this.g("DataToolTipValueFormatMinFractions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueFormatMode", {
        get: function () {
            return this.lz;
        },
        set: function (a) {
            this.lz = a;
            this.g("DataToolTipValueFormatMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueFormatCulture", {
        get: function () {
            return this.ly;
        },
        set: function (a) {
            this.ly = a;
            this.g("DataToolTipValueFormatCulture");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueFormatUseGrouping", {
        get: function () {
            return this.b4;
        },
        set: function (a) {
            this.b4 = a;
            this.g("DataToolTipValueFormatUseGrouping");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueFormatString", {
        get: function () {
            return this.l0;
        },
        set: function (a) {
            this.l0 = a;
            this.g("DataToolTipValueFormatString");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueFormatSpecifiers", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("DataToolTipValueFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueRowMarginBottom", {
        get: function () {
            return this.ex;
        },
        set: function (a) {
            this.ex = a;
            this.g("DataToolTipValueRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueRowMarginLeft", {
        get: function () {
            return this.ey;
        },
        set: function (a) {
            this.ey = a;
            this.g("DataToolTipValueRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueRowMarginRight", {
        get: function () {
            return this.ez;
        },
        set: function (a) {
            this.ez = a;
            this.g("DataToolTipValueRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueRowMarginTop", {
        get: function () {
            return this.e0;
        },
        set: function (a) {
            this.e0 = a;
            this.g("DataToolTipValueRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueRowVisible", {
        get: function () {
            return this.b5;
        },
        set: function (a) {
            this.b5 = a;
            this.g("DataToolTipValueRowVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextWhenMissingData", {
        get: function () {
            return this.l5;
        },
        set: function (a) {
            this.l5 = a;
            this.g("DataToolTipValueTextWhenMissingData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextUseSeriesColors", {
        get: function () {
            return this.b6;
        },
        set: function (a) {
            this.b6 = a;
            this.g("DataToolTipValueTextUseSeriesColors");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextMarginBottom", {
        get: function () {
            return this.e2;
        },
        set: function (a) {
            this.e2 = a;
            this.g("DataToolTipValueTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextMarginLeft", {
        get: function () {
            return this.e3;
        },
        set: function (a) {
            this.e3 = a;
            this.g("DataToolTipValueTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextMarginRight", {
        get: function () {
            return this.e4;
        },
        set: function (a) {
            this.e4 = a;
            this.g("DataToolTipValueTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextMarginTop", {
        get: function () {
            return this.e5;
        },
        set: function (a) {
            this.e5 = a;
            this.g("DataToolTipValueTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextColor", {
        get: function () {
            return this.l1;
        },
        set: function (a) {
            this.l1 = a;
            this.g("DataToolTipValueTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextFontFamily", {
        get: function () {
            return this.l2;
        },
        set: function (a) {
            this.l2 = a;
            this.g("DataToolTipValueTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextFontSize", {
        get: function () {
            return this.e1;
        },
        set: function (a) {
            this.e1 = a;
            this.g("DataToolTipValueTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextFontStyle", {
        get: function () {
            return this.l3;
        },
        set: function (a) {
            this.l3 = a;
            this.g("DataToolTipValueTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipValueTextFontWeight", {
        get: function () {
            return this.l4;
        },
        set: function (a) {
            this.l4 = a;
            this.g("DataToolTipValueTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderFormatString", {
        get: function () {
            return this.kr;
        },
        set: function (a) {
            this.kr = a;
            this.g("DataToolTipHeaderFormatString");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderFormatSpecifiers", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("DataToolTipHeaderFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderFormatCulture", {
        get: function () {
            return this.kp;
        },
        set: function (a) {
            this.kp = a;
            this.g("DataToolTipHeaderFormatCulture");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderFormatDate", {
        get: function () {
            return this.kq;
        },
        set: function (a) {
            this.kq = a;
            this.g("DataToolTipHeaderFormatDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderFormatTime", {
        get: function () {
            return this.ks;
        },
        set: function (a) {
            this.ks = a;
            this.g("DataToolTipHeaderFormatTime");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderText", {
        get: function () {
            return this.kt;
        },
        set: function (a) {
            this.kt = a;
            this.g("DataToolTipHeaderText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderTextColor", {
        get: function () {
            return this.ku;
        },
        set: function (a) {
            this.ku = a;
            this.g("DataToolTipHeaderTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderTextMarginBottom", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.g("DataToolTipHeaderTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderTextMarginLeft", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            this.d1 = a;
            this.g("DataToolTipHeaderTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderTextMarginRight", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            this.d2 = a;
            this.g("DataToolTipHeaderTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderTextMarginTop", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.g("DataToolTipHeaderTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderRowMarginBottom", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.g("DataToolTipHeaderRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderRowMarginLeft", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.g("DataToolTipHeaderRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderRowMarginRight", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("DataToolTipHeaderRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderRowMarginTop", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("DataToolTipHeaderRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderRowVisible", {
        get: function () {
            return this.b2;
        },
        set: function (a) {
            this.b2 = a;
            this.g("DataToolTipHeaderRowVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderTextFontFamily", {
        get: function () {
            return this.kv;
        },
        set: function (a) {
            this.kv = a;
            this.g("DataToolTipHeaderTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderTextFontSize", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.g("DataToolTipHeaderTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderTextFontStyle", {
        get: function () {
            return this.kw;
        },
        set: function (a) {
            this.kw = a;
            this.g("DataToolTipHeaderTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipHeaderTextFontWeight", {
        get: function () {
            return this.kx;
        },
        set: function (a) {
            this.kx = a;
            this.g("DataToolTipHeaderTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupTextColor", {
        get: function () {
            return this.kl;
        },
        set: function (a) {
            this.kl = a;
            this.g("DataToolTipGroupTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupTextMarginBottom", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("DataToolTipGroupTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupTextMarginLeft", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("DataToolTipGroupTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupTextMarginRight", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("DataToolTipGroupTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupTextMarginTop", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.g("DataToolTipGroupTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupRowMarginBottom", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            this.dl = a;
            this.g("DataToolTipGroupRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupRowMarginLeft", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            this.dm = a;
            this.g("DataToolTipGroupRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupRowMarginRight", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.g("DataToolTipGroupRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupRowMarginTop", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("DataToolTipGroupRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupRowVisible", {
        get: function () {
            return this.b1;
        },
        set: function (a) {
            this.b1 = a;
            this.g("DataToolTipGroupRowVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupTextFontFamily", {
        get: function () {
            return this.km;
        },
        set: function (a) {
            this.km = a;
            this.g("DataToolTipGroupTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupTextFontSize", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("DataToolTipGroupTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupTextFontStyle", {
        get: function () {
            return this.kn;
        },
        set: function (a) {
            this.kn = a;
            this.g("DataToolTipGroupTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipGroupTextFontWeight", {
        get: function () {
            return this.ko;
        },
        set: function (a) {
            this.ko = a;
            this.g("DataToolTipGroupTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryTitleTextColor", {
        get: function () {
            return this.k9;
        },
        set: function (a) {
            this.k9 = a;
            this.g("DataToolTipSummaryTitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryTitleTextFontFamily", {
        get: function () {
            return this.la;
        },
        set: function (a) {
            this.la = a;
            this.g("DataToolTipSummaryTitleTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryTitleTextFontSize", {
        get: function () {
            return this.eg;
        },
        set: function (a) {
            this.eg = a;
            this.g("DataToolTipSummaryTitleTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryTitleTextFontStyle", {
        get: function () {
            return this.lb;
        },
        set: function (a) {
            this.lb = a;
            this.g("DataToolTipSummaryTitleTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryTitleTextFontWeight", {
        get: function () {
            return this.lc;
        },
        set: function (a) {
            this.lc = a;
            this.g("DataToolTipSummaryTitleTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryType", {
        get: function () {
            return this.ld;
        },
        set: function (a) {
            this.ld = a;
            this.g("DataToolTipSummaryType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryTitleText", {
        get: function () {
            return this.k8;
        },
        set: function (a) {
            this.k8 = a;
            this.g("DataToolTipSummaryTitleText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryTitleTextMarginBottom", {
        get: function () {
            return this.eh;
        },
        set: function (a) {
            this.eh = a;
            this.g("DataToolTipSummaryTitleTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryTitleTextMarginLeft", {
        get: function () {
            return this.ei;
        },
        set: function (a) {
            this.ei = a;
            this.g("DataToolTipSummaryTitleTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryTitleTextMarginRight", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("DataToolTipSummaryTitleTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryTitleTextMarginTop", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.g("DataToolTipSummaryTitleTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryRowMarginBottom", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            this.ec = a;
            this.g("DataToolTipSummaryRowMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryRowMarginLeft", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            this.ed = a;
            this.g("DataToolTipSummaryRowMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryRowMarginRight", {
        get: function () {
            return this.ee;
        },
        set: function (a) {
            this.ee = a;
            this.g("DataToolTipSummaryRowMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryRowMarginTop", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            this.ef = a;
            this.g("DataToolTipSummaryRowMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryValueTextColor", {
        get: function () {
            return this.lj;
        },
        set: function (a) {
            this.lj = a;
            this.g("DataToolTipSummaryValueTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryValueTextFontFamily", {
        get: function () {
            return this.lk;
        },
        set: function (a) {
            this.lk = a;
            this.g("DataToolTipSummaryValueTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryValueTextFontSize", {
        get: function () {
            return this.em;
        },
        set: function (a) {
            this.em = a;
            this.g("DataToolTipSummaryValueTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryValueTextFontStyle", {
        get: function () {
            return this.ll;
        },
        set: function (a) {
            this.ll = a;
            this.g("DataToolTipSummaryValueTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryValueTextFontWeight", {
        get: function () {
            return this.lm;
        },
        set: function (a) {
            this.lm = a;
            this.g("DataToolTipSummaryValueTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryLabelText", {
        get: function () {
            return this.k3;
        },
        set: function (a) {
            this.k3 = a;
            this.g("DataToolTipSummaryLabelText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryLabelTextColor", {
        get: function () {
            return this.k4;
        },
        set: function (a) {
            this.k4 = a;
            this.g("DataToolTipSummaryLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryLabelTextFontFamily", {
        get: function () {
            return this.k5;
        },
        set: function (a) {
            this.k5 = a;
            this.g("DataToolTipSummaryLabelTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryLabelTextFontSize", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("DataToolTipSummaryLabelTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryLabelTextFontStyle", {
        get: function () {
            return this.k6;
        },
        set: function (a) {
            this.k6 = a;
            this.g("DataToolTipSummaryLabelTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryLabelTextFontWeight", {
        get: function () {
            return this.k7;
        },
        set: function (a) {
            this.k7 = a;
            this.g("DataToolTipSummaryLabelTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryUnitsText", {
        get: function () {
            return this.le;
        },
        set: function (a) {
            this.le = a;
            this.g("DataToolTipSummaryUnitsText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryUnitsTextColor", {
        get: function () {
            return this.lf;
        },
        set: function (a) {
            this.lf = a;
            this.g("DataToolTipSummaryUnitsTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryUnitsTextFontFamily", {
        get: function () {
            return this.lg;
        },
        set: function (a) {
            this.lg = a;
            this.g("DataToolTipSummaryUnitsTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryUnitsTextFontSize", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            this.el = a;
            this.g("DataToolTipSummaryUnitsTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryUnitsTextFontStyle", {
        get: function () {
            return this.lh;
        },
        set: function (a) {
            this.lh = a;
            this.g("DataToolTipSummaryUnitsTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipSummaryUnitsTextFontWeight", {
        get: function () {
            return this.li;
        },
        set: function (a) {
            this.li = a;
            this.g("DataToolTipSummaryUnitsTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipBadgeMarginBottom", {
        get: function () {
            return this.df;
        },
        set: function (a) {
            this.df = a;
            this.g("DataToolTipBadgeMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipBadgeMarginLeft", {
        get: function () {
            return this.dg;
        },
        set: function (a) {
            this.dg = a;
            this.g("DataToolTipBadgeMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipBadgeMarginRight", {
        get: function () {
            return this.dh;
        },
        set: function (a) {
            this.dh = a;
            this.g("DataToolTipBadgeMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipBadgeMarginTop", {
        get: function () {
            return this.di;
        },
        set: function (a) {
            this.di = a;
            this.g("DataToolTipBadgeMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipBadgeShape", {
        get: function () {
            return this.kh;
        },
        set: function (a) {
            this.kh = a;
            this.g("DataToolTipBadgeShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsDisplayMode", {
        get: function () {
            return this.lr;
        },
        set: function (a) {
            this.lr = a;
            this.g("DataToolTipUnitsDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsText", {
        get: function () {
            return this.ls;
        },
        set: function (a) {
            this.ls = a;
            this.g("DataToolTipUnitsText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsTextColor", {
        get: function () {
            return this.lt;
        },
        set: function (a) {
            this.lt = a;
            this.g("DataToolTipUnitsTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsTextMarginBottom", {
        get: function () {
            return this.et;
        },
        set: function (a) {
            this.et = a;
            this.g("DataToolTipUnitsTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsTextMarginLeft", {
        get: function () {
            return this.eu;
        },
        set: function (a) {
            this.eu = a;
            this.g("DataToolTipUnitsTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsTextMarginRight", {
        get: function () {
            return this.ev;
        },
        set: function (a) {
            this.ev = a;
            this.g("DataToolTipUnitsTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsTextMarginTop", {
        get: function () {
            return this.ew;
        },
        set: function (a) {
            this.ew = a;
            this.g("DataToolTipUnitsTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsTextFontFamily", {
        get: function () {
            return this.lu;
        },
        set: function (a) {
            this.lu = a;
            this.g("DataToolTipUnitsTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsTextFontSize", {
        get: function () {
            return this.es;
        },
        set: function (a) {
            this.es = a;
            this.g("DataToolTipUnitsTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsTextFontStyle", {
        get: function () {
            return this.lv;
        },
        set: function (a) {
            this.lv = a;
            this.g("DataToolTipUnitsTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipUnitsTextFontWeight", {
        get: function () {
            return this.lw;
        },
        set: function (a) {
            this.lw = a;
            this.g("DataToolTipUnitsTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipTitleTextMarginBottom", {
        get: function () {
            return this.eo;
        },
        set: function (a) {
            this.eo = a;
            this.g("DataToolTipTitleTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipTitleTextMarginLeft", {
        get: function () {
            return this.ep;
        },
        set: function (a) {
            this.ep = a;
            this.g("DataToolTipTitleTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipTitleTextMarginRight", {
        get: function () {
            return this.eq;
        },
        set: function (a) {
            this.eq = a;
            this.g("DataToolTipTitleTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipTitleTextMarginTop", {
        get: function () {
            return this.er;
        },
        set: function (a) {
            this.er = a;
            this.g("DataToolTipTitleTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipTitleTextColor", {
        get: function () {
            return this.ln;
        },
        set: function (a) {
            this.ln = a;
            this.g("DataToolTipTitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipTitleTextFontFamily", {
        get: function () {
            return this.lo;
        },
        set: function (a) {
            this.lo = a;
            this.g("DataToolTipTitleTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipTitleTextFontSize", {
        get: function () {
            return this.en;
        },
        set: function (a) {
            this.en = a;
            this.g("DataToolTipTitleTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipTitleTextFontStyle", {
        get: function () {
            return this.lp;
        },
        set: function (a) {
            this.lp = a;
            this.g("DataToolTipTitleTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipTitleTextFontWeight", {
        get: function () {
            return this.lq;
        },
        set: function (a) {
            this.lq = a;
            this.g("DataToolTipTitleTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipLabelDisplayMode", {
        get: function () {
            return this.ky;
        },
        set: function (a) {
            this.ky = a;
            this.g("DataToolTipLabelDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipLabelTextColor", {
        get: function () {
            return this.kz;
        },
        set: function (a) {
            this.kz = a;
            this.g("DataToolTipLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipLabelTextMarginBottom", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("DataToolTipLabelTextMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipLabelTextMarginLeft", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            this.d6 = a;
            this.g("DataToolTipLabelTextMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipLabelTextMarginRight", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("DataToolTipLabelTextMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipLabelTextMarginTop", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.g("DataToolTipLabelTextMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipLabelTextFontFamily", {
        get: function () {
            return this.k0;
        },
        set: function (a) {
            this.k0 = a;
            this.g("DataToolTipLabelTextFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipLabelTextFontSize", {
        get: function () {
            return this.d4;
        },
        set: function (a) {
            this.d4 = a;
            this.g("DataToolTipLabelTextFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipLabelTextFontStyle", {
        get: function () {
            return this.k1;
        },
        set: function (a) {
            this.k1 = a;
            this.g("DataToolTipLabelTextFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataToolTipLabelTextFontWeight", {
        get: function () {
            return this.k2;
        },
        set: function (a) {
            this.k2 = a;
            this.g("DataToolTipLabelTextFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "sortDescriptions", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("SortDescriptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "groupSortDescriptions", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("GroupSortDescriptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "groupDescriptions", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("GroupDescriptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "filterExpressions", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("FilterExpressions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "highlightFilterExpressions", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("HighlightFilterExpressions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "summaryDescriptions", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("SummaryDescriptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "selectionMode", {
        get: function () {
            return this.m8;
        },
        set: function (a) {
            this.m8 = a;
            this.g("SelectionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "focusMode", {
        get: function () {
            return this.mb;
        },
        set: function (a) {
            this.mb = a;
            this.g("FocusMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "selectionBrush", {
        get: function () {
            return this.m7;
        },
        set: function (a) {
            this.m7 = a;
            this.g("SelectionBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "focusBrush", {
        get: function () {
            return this.l9;
        },
        set: function (a) {
            this.l9 = a;
            this.g("FocusBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "selectionBehavior", {
        get: function () {
            return this.m6;
        },
        set: function (a) {
            this.m6 = a;
            this.g("SelectionBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "selectedSeriesItems", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("SelectedSeriesItems");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "focusedSeriesItems", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("FocusedSeriesItems");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialSortDescriptions", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("InitialSortDescriptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialGroupSortDescriptions", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("InitialGroupSortDescriptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialGroupDescriptions", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("InitialGroupDescriptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialFilterExpressions", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("InitialFilterExpressions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialHighlightFilterExpressions", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("InitialHighlightFilterExpressions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialSummaryDescriptions", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("InitialSummaryDescriptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialSorts", {
        get: function () {
            return this.mp;
        },
        set: function (a) {
            this.mp = a;
            this.g("InitialSorts");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "groupSorts", {
        get: function () {
            return this.mc;
        },
        set: function (a) {
            this.mc = a;
            this.g("GroupSorts");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialGroups", {
        get: function () {
            return this.mn;
        },
        set: function (a) {
            this.mn = a;
            this.g("InitialGroups");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialFilter", {
        get: function () {
            return this.mm;
        },
        set: function (a) {
            this.mm = a;
            this.g("InitialFilter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialHighlightFilter", {
        get: function () {
            return this.mo;
        },
        set: function (a) {
            this.mo = a;
            this.g("InitialHighlightFilter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "initialSummaries", {
        get: function () {
            return this.mq;
        },
        set: function (a) {
            this.mq = a;
            this.g("InitialSummaries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.kg;
        },
        set: function (a) {
            this.kg = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "dataSource", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.g("DataSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "highlightedDataSourceRef", {
        get: function () {
            return this.md;
        },
        set: function (a) {
            this.md = a;
            this.g("HighlightedDataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "highlightedDataSource", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            this.bm = a;
            this.g("HighlightedDataSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "includedProperties", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("IncludedProperties");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "excludedProperties", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("ExcludedProperties");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "brushes", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Brushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "outlines", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("Outlines");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "actualBrushes", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ActualBrushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "actualOutlines", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ActualOutlines");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "legendRef", {
        get: function () {
            return this.mv;
        },
        set: function (a) {
            this.mv = a;
            this.g("LegendRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "isHorizontalZoomEnabled", {
        get: function () {
            return this.b9;
        },
        set: function (a) {
            this.b9 = a;
            this.g("IsHorizontalZoomEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "isVerticalZoomEnabled", {
        get: function () {
            return this.cb;
        },
        set: function (a) {
            this.cb = a;
            this.g("IsVerticalZoomEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "shouldDisplayMockData", {
        get: function () {
            return this.ce;
        },
        set: function (a) {
            this.ce = a;
            this.g("ShouldDisplayMockData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "shouldUseSkeletonStyleForMockData", {
        get: function () {
            return this.cg;
        },
        set: function (a) {
            this.cg = a;
            this.g("ShouldUseSkeletonStyleForMockData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "highlightedValuesDisplayMode", {
        get: function () {
            return this.mf;
        },
        set: function (a) {
            this.mf = a;
            this.g("HighlightedValuesDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "isSeriesHighlightingEnabled", {
        get: function () {
            return this.ca;
        },
        set: function (a) {
            this.ca = a;
            this.g("IsSeriesHighlightingEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "highlightedLegendItemVisibility", {
        get: function () {
            return this.me;
        },
        set: function (a) {
            this.me = a;
            this.g("HighlightedLegendItemVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "legendItemVisibility", {
        get: function () {
            return this.mu;
        },
        set: function (a) {
            this.mu = a;
            this.g("LegendItemVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "windowRect", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.g("WindowRect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "chartTitle", {
        get: function () {
            return this.j7;
        },
        set: function (a) {
            this.j7 = a;
            this.g("ChartTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "subtitle", {
        get: function () {
            return this.nj;
        },
        set: function (a) {
            this.nj = a;
            this.g("Subtitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "titleAlignment", {
        get: function () {
            return this.nn;
        },
        set: function (a) {
            this.nn = a;
            this.g("TitleAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "subtitleAlignment", {
        get: function () {
            return this.nk;
        },
        set: function (a) {
            this.nk = a;
            this.g("SubtitleAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "unknownValuePlotting", {
        get: function () {
            return this.nt;
        },
        set: function (a) {
            this.nt = a;
            this.g("UnknownValuePlotting");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "resolution", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("Resolution");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "thickness", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("Thickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "outlineMode", {
        get: function () {
            return this.mz;
        },
        set: function (a) {
            this.mz = a;
            this.g("OutlineMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "markerOutlineMode", {
        get: function () {
            return this.my;
        },
        set: function (a) {
            this.my = a;
            this.g("MarkerOutlineMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "markerFillMode", {
        get: function () {
            return this.mx;
        },
        set: function (a) {
            this.mx = a;
            this.g("MarkerFillMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "markerFillOpacity", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("MarkerFillOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "markerThickness", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            this.ff = a;
            this.g("MarkerThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "markerTypes", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("MarkerTypes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "markerAutomaticBehavior", {
        get: function () {
            return this.mw;
        },
        set: function (a) {
            this.mw = a;
            this.g("MarkerAutomaticBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "markerBrushes", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("MarkerBrushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "markerOutlines", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("MarkerOutlines");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "markerMaxCount", {
        get: function () {
            return this.ji;
        },
        set: function (a) {
            this.ji = a;
            this.g("MarkerMaxCount");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "areaFillOpacity", {
        get: function () {
            return this.db;
        },
        set: function (a) {
            this.db = a;
            this.g("AreaFillOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "animateSeriesWhenAxisRangeChanges", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("AnimateSeriesWhenAxisRangeChanges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "trendLineBrushes", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("TrendLineBrushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "trendLineType", {
        get: function () {
            return this.ns;
        },
        set: function (a) {
            this.ns = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "alignsGridLinesToPixels", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.g("AlignsGridLinesToPixels");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "plotAreaMarginLeft", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("PlotAreaMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "plotAreaMarginTop", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("PlotAreaMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "plotAreaMarginRight", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("PlotAreaMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "plotAreaMarginBottom", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("PlotAreaMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "computedPlotAreaMarginMode", {
        get: function () {
            return this.j8;
        },
        set: function (a) {
            this.j8 = a;
            this.g("ComputedPlotAreaMarginMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesPlotAreaMarginHorizontalMode", {
        get: function () {
            return this.nb;
        },
        set: function (a) {
            this.nb = a;
            this.g("SeriesPlotAreaMarginHorizontalMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesPlotAreaMarginVerticalMode", {
        get: function () {
            return this.nc;
        },
        set: function (a) {
            this.nc = a;
            this.g("SeriesPlotAreaMarginVerticalMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "highlightingMode", {
        get: function () {
            return this.mh;
        },
        set: function (a) {
            this.mh = a;
            this.g("HighlightingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "highlightingBehavior", {
        get: function () {
            return this.mg;
        },
        set: function (a) {
            this.mg = a;
            this.g("HighlightingBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "legendHighlightingMode", {
        get: function () {
            return this.mr;
        },
        set: function (a) {
            this.mr = a;
            this.g("LegendHighlightingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "legendItemBadgeShape", {
        get: function () {
            return this.mt;
        },
        set: function (a) {
            this.mt = a;
            this.g("LegendItemBadgeShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "legendItemBadgeMode", {
        get: function () {
            return this.ms;
        },
        set: function (a) {
            this.ms = a;
            this.g("LegendItemBadgeMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.jl;
        },
        set: function (a) {
            this.jl = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "toolTipType", {
        get: function () {
            return this.nq;
        },
        set: function (a) {
            this.nq = a;
            this.g("ToolTipType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsDisplayMode", {
        get: function () {
            return this.kd;
        },
        set: function (a) {
            this.kd = a;
            this.g("CrosshairsDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsSnapToData", {
        get: function () {
            return this.b0;
        },
        set: function (a) {
            this.b0 = a;
            this.g("CrosshairsSnapToData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsLineVerticalStroke", {
        get: function () {
            return this.kf;
        },
        set: function (a) {
            this.kf = a;
            this.g("CrosshairsLineVerticalStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsLineThickness", {
        get: function () {
            return this.de;
        },
        set: function (a) {
            this.de = a;
            this.g("CrosshairsLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsLineHorizontalStroke", {
        get: function () {
            return this.ke;
        },
        set: function (a) {
            this.ke = a;
            this.g("CrosshairsLineHorizontalStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsAnnotationEnabled", {
        get: function () {
            return this.bz;
        },
        set: function (a) {
            this.bz = a;
            this.g("CrosshairsAnnotationEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsAnnotationXAxisBackground", {
        get: function () {
            return this.j9;
        },
        set: function (a) {
            this.j9 = a;
            this.g("CrosshairsAnnotationXAxisBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsAnnotationYAxisBackground", {
        get: function () {
            return this.kb;
        },
        set: function (a) {
            this.kb = a;
            this.g("CrosshairsAnnotationYAxisBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsAnnotationXAxisTextColor", {
        get: function () {
            return this.ka;
        },
        set: function (a) {
            this.ka = a;
            this.g("CrosshairsAnnotationXAxisTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsAnnotationYAxisTextColor", {
        get: function () {
            return this.kc;
        },
        set: function (a) {
            this.kc = a;
            this.g("CrosshairsAnnotationYAxisTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsAnnotationXAxisPrecision", {
        get: function () {
            return this.jb;
        },
        set: function (a) {
            this.jb = a;
            this.g("CrosshairsAnnotationXAxisPrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "crosshairsAnnotationYAxisPrecision", {
        get: function () {
            return this.jc;
        },
        set: function (a) {
            this.jc = a;
            this.g("CrosshairsAnnotationYAxisPrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "shouldAvoidAxisAnnotationCollisions", {
        get: function () {
            return this.cd;
        },
        set: function (a) {
            this.cd = a;
            this.g("ShouldAvoidAxisAnnotationCollisions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "shouldPanOnMaximumZoom", {
        get: function () {
            return this.cf;
        },
        set: function (a) {
            this.cf = a;
            this.g("ShouldPanOnMaximumZoom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "finalValueAnnotationsVisible", {
        get: function () {
            return this.b7;
        },
        set: function (a) {
            this.b7 = a;
            this.g("FinalValueAnnotationsVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "finalValueAnnotationsBackground", {
        get: function () {
            return this.l7;
        },
        set: function (a) {
            this.l7 = a;
            this.g("FinalValueAnnotationsBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "finalValueAnnotationsTextColor", {
        get: function () {
            return this.l8;
        },
        set: function (a) {
            this.l8 = a;
            this.g("FinalValueAnnotationsTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "finalValueAnnotationsPrecision", {
        get: function () {
            return this.jf;
        },
        set: function (a) {
            this.jf = a;
            this.g("FinalValueAnnotationsPrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "autoCalloutsVisible", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("AutoCalloutsVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsVisible", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            this.by = a;
            this.g("CalloutsVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutStyleUpdatingEventEnabled", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("CalloutStyleUpdatingEventEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "useValueForAutoCalloutLabels", {
        get: function () {
            return this.ch;
        },
        set: function (a) {
            this.ch = a;
            this.g("UseValueForAutoCalloutLabels");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutCollisionMode", {
        get: function () {
            return this.jy;
        },
        set: function (a) {
            this.jy = a;
            this.g("CalloutCollisionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsUseItemColorForFill", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("CalloutsUseItemColorForFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsUseItemColorForOutline", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("CalloutsUseItemColorForOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsUseAutoContrastingLabelColors", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("CalloutsUseAutoContrastingLabelColors");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsStrokeThickness", {
        get: function () {
            return this.dd;
        },
        set: function (a) {
            this.dd = a;
            this.g("CalloutsStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsBackground", {
        get: function () {
            return this.j0;
        },
        set: function (a) {
            this.j0 = a;
            this.g("CalloutsBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsAutoLabelPrecision", {
        get: function () {
            return this.ja;
        },
        set: function (a) {
            this.ja = a;
            this.g("CalloutsAutoLabelPrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsDataSourceRef", {
        get: function () {
            return this.j2;
        },
        set: function (a) {
            this.j2 = a;
            this.g("CalloutsDataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsAllowedPositions", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("CalloutsAllowedPositions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsXMemberPath", {
        get: function () {
            return this.j5;
        },
        set: function (a) {
            this.j5 = a;
            this.g("CalloutsXMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsYMemberPath", {
        get: function () {
            return this.j6;
        },
        set: function (a) {
            this.j6 = a;
            this.g("CalloutsYMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsLabelMemberPath", {
        get: function () {
            return this.j3;
        },
        set: function (a) {
            this.j3 = a;
            this.g("CalloutsLabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutsContentMemberPath", {
        get: function () {
            return this.j1;
        },
        set: function (a) {
            this.j1 = a;
            this.g("CalloutsContentMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "valueLines", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("ValueLines");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "valueLinesBrushes", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("ValueLinesBrushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "valueLinesThickness", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            this.fy = a;
            this.g("ValueLinesThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesValueLayerUseLegend", {
        get: function () {
            return this.cc;
        },
        set: function (a) {
            this.cc = a;
            this.g("SeriesValueLayerUseLegend");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarMode", {
        get: function () {
            return this.mj;
        },
        set: function (a) {
            this.mj = a;
            this.g("HorizontalViewScrollbarMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarMode", {
        get: function () {
            return this.nv;
        },
        set: function (a) {
            this.nv = a;
            this.g("VerticalViewScrollbarMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarPosition", {
        get: function () {
            return this.ml;
        },
        set: function (a) {
            this.ml = a;
            this.g("HorizontalViewScrollbarPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarPosition", {
        get: function () {
            return this.nx;
        },
        set: function (a) {
            this.nx = a;
            this.g("VerticalViewScrollbarPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarFill", {
        get: function () {
            return this.mi;
        },
        set: function (a) {
            this.mi = a;
            this.g("HorizontalViewScrollbarFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarOutline", {
        get: function () {
            return this.mk;
        },
        set: function (a) {
            this.mk = a;
            this.g("HorizontalViewScrollbarOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarStrokeThickness", {
        get: function () {
            return this.fa;
        },
        set: function (a) {
            this.fa = a;
            this.g("HorizontalViewScrollbarStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarMaxOpacity", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("HorizontalViewScrollbarMaxOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarCornerRadius", {
        get: function () {
            return this.e6;
        },
        set: function (a) {
            this.e6 = a;
            this.g("HorizontalViewScrollbarCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarHeight", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("HorizontalViewScrollbarHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarInset", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("HorizontalViewScrollbarInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarTrackStartInset", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("HorizontalViewScrollbarTrackStartInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarTrackEndInset", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("HorizontalViewScrollbarTrackEndInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "horizontalViewScrollbarShouldAddAutoTrackInsets", {
        get: function () {
            return this.b8;
        },
        set: function (a) {
            this.b8 = a;
            this.g("HorizontalViewScrollbarShouldAddAutoTrackInsets");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarFill", {
        get: function () {
            return this.nu;
        },
        set: function (a) {
            this.nu = a;
            this.g("VerticalViewScrollbarFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarOutline", {
        get: function () {
            return this.nw;
        },
        set: function (a) {
            this.nw = a;
            this.g("VerticalViewScrollbarOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarStrokeThickness", {
        get: function () {
            return this.f2;
        },
        set: function (a) {
            this.f2 = a;
            this.g("VerticalViewScrollbarStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarMaxOpacity", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("VerticalViewScrollbarMaxOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarCornerRadius", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("VerticalViewScrollbarCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarWidth", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("VerticalViewScrollbarWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarInset", {
        get: function () {
            return this.f0;
        },
        set: function (a) {
            this.f0 = a;
            this.g("VerticalViewScrollbarInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarTrackStartInset", {
        get: function () {
            return this.f4;
        },
        set: function (a) {
            this.f4 = a;
            this.g("VerticalViewScrollbarTrackStartInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarTrackEndInset", {
        get: function () {
            return this.f3;
        },
        set: function (a) {
            this.f3 = a;
            this.g("VerticalViewScrollbarTrackEndInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "verticalViewScrollbarShouldAddAutoTrackInsets", {
        get: function () {
            return this.ci;
        },
        set: function (a) {
            this.ci = a;
            this.g("VerticalViewScrollbarShouldAddAutoTrackInsets");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "windowRectMinWidth", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("WindowRectMinWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "windowRectMinHeight", {
        get: function () {
            return this.f6;
        },
        set: function (a) {
            this.f6 = a;
            this.g("WindowRectMinHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "windowSizeMinWidth", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("WindowSizeMinWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "windowSizeMinHeight", {
        get: function () {
            return this.f8;
        },
        set: function (a) {
            this.f8 = a;
            this.g("WindowSizeMinHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesAddedRef", {
        get: function () {
            return this.m9;
        },
        set: function (a) {
            this.m9 = a;
            this.g("SeriesAddedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesRemovedRef", {
        get: function () {
            return this.ni;
        },
        set: function (a) {
            this.ni = a;
            this.g("SeriesRemovedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesPointerEnterRef", {
        get: function () {
            return this.ne;
        },
        set: function (a) {
            this.ne = a;
            this.g("SeriesPointerEnterRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesPointerLeaveRef", {
        get: function () {
            return this.nf;
        },
        set: function (a) {
            this.nf = a;
            this.g("SeriesPointerLeaveRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesPointerMoveRef", {
        get: function () {
            return this.ng;
        },
        set: function (a) {
            this.ng = a;
            this.g("SeriesPointerMoveRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesPointerDownRef", {
        get: function () {
            return this.nd;
        },
        set: function (a) {
            this.nd = a;
            this.g("SeriesPointerDownRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesPointerUpRef", {
        get: function () {
            return this.nh;
        },
        set: function (a) {
            this.nh = a;
            this.g("SeriesPointerUpRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "seriesClickRef", {
        get: function () {
            return this.na;
        },
        set: function (a) {
            this.na = a;
            this.g("SeriesClickRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "plotAreaPointerEnterRef", {
        get: function () {
            return this.m1;
        },
        set: function (a) {
            this.m1 = a;
            this.g("PlotAreaPointerEnterRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "plotAreaPointerLeaveRef", {
        get: function () {
            return this.m2;
        },
        set: function (a) {
            this.m2 = a;
            this.g("PlotAreaPointerLeaveRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "plotAreaPointerMoveRef", {
        get: function () {
            return this.m3;
        },
        set: function (a) {
            this.m3 = a;
            this.g("PlotAreaPointerMoveRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "plotAreaPointerDownRef", {
        get: function () {
            return this.m0;
        },
        set: function (a) {
            this.m0 = a;
            this.g("PlotAreaPointerDownRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "plotAreaPointerUpRef", {
        get: function () {
            return this.m4;
        },
        set: function (a) {
            this.m4 = a;
            this.g("PlotAreaPointerUpRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutStyleUpdatingRef", {
        get: function () {
            return this.j4;
        },
        set: function (a) {
            this.j4 = a;
            this.g("CalloutStyleUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "calloutRenderStyleUpdatingRef", {
        get: function () {
            return this.jz;
        },
        set: function (a) {
            this.jz = a;
            this.g("CalloutRenderStyleUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "selectedSeriesItemsChangedRef", {
        get: function () {
            return this.m5;
        },
        set: function (a) {
            this.m5 = a;
            this.g("SelectedSeriesItemsChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "focusedSeriesItemsChangedRef", {
        get: function () {
            return this.ma;
        },
        set: function (a) {
            this.ma = a;
            this.g("FocusedSeriesItemsChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomainChartDescription.prototype, "filterStringErrorsParsingRef", {
        get: function () {
            return this.l6;
        },
        set: function (a) {
            this.l6 = a;
            this.g("FilterStringErrorsParsingRef");
        },
        enumerable: false,
        configurable: true
    });
    DomainChartDescription.$t = markType(DomainChartDescription, 'DomainChartDescription', Description.$);
    return DomainChartDescription;
}(Description));
export { DomainChartDescription };
