// igniteui-react-core/esm5/lib/EditorCellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { CellInfoDescription } from "./CellInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var EditorCellInfoDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(EditorCellInfoDescription, _super);
    function EditorCellInfoDescription() {
        var _this = _super.call(this) || this;
        _this.fn = null;
        _this.ft = null;
        _this.fz = null;
        _this.fy = null;
        _this.fx = null;
        _this.fv = null;
        _this.fw = null;
        _this.fl = null;
        _this.fu = null;
        _this.fq = false;
        return _this;
    }
    EditorCellInfoDescription.prototype.get_type = function () {
        return "EditorCellInfo";
    };
    Object.defineProperty(EditorCellInfoDescription.prototype, "editTarget", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("EditTarget");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EditorCellInfoDescription.prototype, "dataType", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("DataType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EditorCellInfoDescription.prototype, "errorMessage", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("ErrorMessage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EditorCellInfoDescription.prototype, "editValueRef", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            this.fy = a;
            this.g("EditValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EditorCellInfoDescription.prototype, "editorType", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("EditorType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EditorCellInfoDescription.prototype, "editorDataSourceRef", {
        get: function () {
            return this.fv;
        },
        set: function (a) {
            this.fv = a;
            this.g("EditorDataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EditorCellInfoDescription.prototype, "editorTextField", {
        get: function () {
            return this.fw;
        },
        set: function (a) {
            this.fw = a;
            this.g("EditorTextField");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EditorCellInfoDescription.prototype, "editorValueField", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("EditorValueField");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EditorCellInfoDescription.prototype, "dateFormatString", {
        get: function () {
            return this.fu;
        },
        set: function (a) {
            this.fu = a;
            this.g("DateFormatString");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EditorCellInfoDescription.prototype, "showTodayButton", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("ShowTodayButton");
        },
        enumerable: false,
        configurable: true
    });
    EditorCellInfoDescription.$t = markType(EditorCellInfoDescription, 'EditorCellInfoDescription', CellInfoDescription.$);
    EditorCellInfoDescription.__marshalByValue1 = true;
    EditorCellInfoDescription.__marshalByValueAlias1 = "EditorCellModel";
    return EditorCellInfoDescription;
}(CellInfoDescription));
export { EditorCellInfoDescription };
