// igniteui-react-core/esm5/lib/FilterRowDefinitionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var FilterRowDefinitionDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FilterRowDefinitionDescription, _super);
    function FilterRowDefinitionDescription() {
        var _this = _super.call(this) || this;
        _this.cq = 0;
        _this.cs = 0;
        _this.cr = 0;
        _this.cp = 0;
        return _this;
    }
    FilterRowDefinitionDescription.prototype.get_type = function () {
        return "FilterRowDefinition";
    };
    Object.defineProperty(FilterRowDefinitionDescription.prototype, "paddingLeft", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("PaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FilterRowDefinitionDescription.prototype, "paddingTop", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("PaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FilterRowDefinitionDescription.prototype, "paddingRight", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("PaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FilterRowDefinitionDescription.prototype, "paddingBottom", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("PaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    FilterRowDefinitionDescription.$t = markType(FilterRowDefinitionDescription, 'FilterRowDefinitionDescription', DefinitionBaseDescription.$);
    FilterRowDefinitionDescription.__canMarshalByValue = true;
    FilterRowDefinitionDescription.__marshalByValueAlias = "FilterRowDefinition";
    return FilterRowDefinitionDescription;
}(DefinitionBaseDescription));
export { FilterRowDefinitionDescription };
