// igniteui-react-core/esm5/lib/FinancialCalculationDataSourceDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var FinancialCalculationDataSourceDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialCalculationDataSourceDescription, _super);
    function FinancialCalculationDataSourceDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.i = null;
        _this.h = null;
        _this.z = 0;
        _this.aa = 0;
        _this.y = 0;
        _this.x = 0;
        _this.w = 0;
        _this.v = 0;
        _this.r = 0;
        _this.q = 0;
        _this.p = 0;
        _this.n = false;
        return _this;
    }
    FinancialCalculationDataSourceDescription.prototype.get_type = function () {
        return "FinancialCalculationDataSource";
    };
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "typicalColumn", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("TypicalColumn");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "trueRange", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("TrueRange");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "trueLow", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("TrueLow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "period", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("Period");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "shortPeriod", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("ShortPeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "longPeriod", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("LongPeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "count", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("Count");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "calculateFrom", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("CalculateFrom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "calculateCount", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("CalculateCount");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "multiplier", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Multiplier");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "minimumValue", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("MinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "maximumValue", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("MaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSourceDescription.prototype, "specifiesRange", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("SpecifiesRange");
        },
        enumerable: false,
        configurable: true
    });
    FinancialCalculationDataSourceDescription.$t = markType(FinancialCalculationDataSourceDescription, 'FinancialCalculationDataSourceDescription', Description.$);
    return FinancialCalculationDataSourceDescription;
}(Description));
export { FinancialCalculationDataSourceDescription };
