// igniteui-react-core/esm5/lib/FinalValueLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var FinalValueLayerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinalValueLayerDescription, _super);
    function FinalValueLayerDescription() {
        var _this = _super.call(this) || this;
        _this.gs = null;
        _this.gt = null;
        _this.gr = null;
        _this.gq = null;
        _this.gn = null;
        _this.f9 = 0;
        _this.gl = 0;
        _this.gp = null;
        _this.gb = 0;
        _this.gd = 0;
        _this.gc = 0;
        _this.ga = 0;
        _this.ge = 0;
        _this.go = null;
        return _this;
    }
    FinalValueLayerDescription.prototype.get_type = function () {
        return "FinalValueLayer";
    };
    Object.defineProperty(FinalValueLayerDescription.prototype, "targetSeriesName", {
        get: function () {
            return this.gs;
        },
        set: function (a) {
            this.gs = a;
            this.g("TargetSeriesName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "targetSeriesRef", {
        get: function () {
            return this.gt;
        },
        set: function (a) {
            this.gt = a;
            this.g("TargetSeriesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "finalValueSelectionMode", {
        get: function () {
            return this.gr;
        },
        set: function (a) {
            this.gr = a;
            this.g("FinalValueSelectionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationTextColor", {
        get: function () {
            return this.gq;
        },
        set: function (a) {
            this.gq = a;
            this.g("AxisAnnotationTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationBackground", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("AxisAnnotationBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationBackgroundCornerRadius", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("AxisAnnotationBackgroundCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationInterpolatedValuePrecision", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("AxisAnnotationInterpolatedValuePrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationOutline", {
        get: function () {
            return this.gp;
        },
        set: function (a) {
            this.gp = a;
            this.g("AxisAnnotationOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationPaddingLeft", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("AxisAnnotationPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationPaddingTop", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("AxisAnnotationPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationPaddingRight", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("AxisAnnotationPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationPaddingBottom", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("AxisAnnotationPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationStrokeThickness", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("AxisAnnotationStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinalValueLayerDescription.prototype, "axisAnnotationFormatLabelRef", {
        get: function () {
            return this.go;
        },
        set: function (a) {
            this.go = a;
            this.g("AxisAnnotationFormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    FinalValueLayerDescription.$t = markType(FinalValueLayerDescription, 'FinalValueLayerDescription', AnnotationLayerDescription.$);
    return FinalValueLayerDescription;
}(AnnotationLayerDescription));
export { FinalValueLayerDescription };
