// igniteui-react-core/esm5/lib/FinancialSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var FinancialSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialSeriesDescription, _super);
    function FinancialSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fg = null;
        _this.fm = null;
        _this.fn = null;
        _this.fh = null;
        _this.fe = null;
        _this.ff = null;
        _this.e8 = null;
        _this.fl = null;
        _this.fc = null;
        _this.fa = null;
        _this.fb = null;
        _this.e9 = null;
        _this.fd = null;
        _this.e3 = false;
        _this.fi = null;
        _this.e4 = false;
        _this.e7 = null;
        _this.fj = null;
        _this.fk = null;
        return _this;
    }
    FinancialSeriesDescription.prototype.get_type = function () {
        return "FinancialSeries";
    };
    Object.defineProperty(FinancialSeriesDescription.prototype, "negativeBrush", {
        get: function () {
            return this.fg;
        },
        set: function (a) {
            this.fg = a;
            this.g("NegativeBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "xAxisRef", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("XAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "yAxisRef", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("YAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "openMemberPath", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("OpenMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "highMemberPath", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("HighMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "lowMemberPath", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            this.ff = a;
            this.g("LowMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "closeMemberPath", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("CloseMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "volumeMemberPath", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("VolumeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "highlightedOpenMemberPath", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("HighlightedOpenMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "highlightedHighMemberPath", {
        get: function () {
            return this.fa;
        },
        set: function (a) {
            this.fa = a;
            this.g("HighlightedHighMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "highlightedLowMemberPath", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("HighlightedLowMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "highlightedCloseMemberPath", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("HighlightedCloseMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "highlightedVolumeMemberPath", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("HighlightedVolumeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "isCustomCategoryStyleAllowed", {
        get: function () {
            return this.e3;
        },
        set: function (a) {
            this.e3 = a;
            this.g("IsCustomCategoryStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "transitionInMode", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("TransitionInMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "isTransitionInEnabled", {
        get: function () {
            return this.e4;
        },
        set: function (a) {
            this.e4 = a;
            this.g("IsTransitionInEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "assigningCategoryStyleRef", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("AssigningCategoryStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "typicalRef", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("TypicalRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeriesDescription.prototype, "typicalBasedOnRef", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("TypicalBasedOnRef");
        },
        enumerable: false,
        configurable: true
    });
    FinancialSeriesDescription.$t = markType(FinancialSeriesDescription, 'FinancialSeriesDescription', SeriesDescription.$);
    return FinancialSeriesDescription;
}(SeriesDescription));
export { FinancialSeriesDescription };
