// igniteui-react-core/esm5/lib/FinancialCalculationSupportingCalculationsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var FinancialCalculationSupportingCalculationsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialCalculationSupportingCalculationsDescription, _super);
    function FinancialCalculationSupportingCalculationsDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.j = null;
        _this.k = null;
        _this.i = null;
        _this.s = null;
        _this.q = null;
        _this.r = null;
        _this.p = null;
        _this.x = null;
        return _this;
    }
    FinancialCalculationSupportingCalculationsDescription.prototype.get_type = function () {
        return "FinancialCalculationSupportingCalculations";
    };
    Object.defineProperty(FinancialCalculationSupportingCalculationsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationSupportingCalculationsDescription.prototype, "eMA", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("EMA");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationSupportingCalculationsDescription.prototype, "sMA", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("SMA");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationSupportingCalculationsDescription.prototype, "sTDEV", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("STDEV");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationSupportingCalculationsDescription.prototype, "movingSum", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("MovingSum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationSupportingCalculationsDescription.prototype, "shortVolumeOscillatorAverage", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("ShortVolumeOscillatorAverage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationSupportingCalculationsDescription.prototype, "longVolumeOscillatorAverage", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("LongVolumeOscillatorAverage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationSupportingCalculationsDescription.prototype, "shortPriceOscillatorAverage", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("ShortPriceOscillatorAverage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationSupportingCalculationsDescription.prototype, "longPriceOscillatorAverage", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("LongPriceOscillatorAverage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationSupportingCalculationsDescription.prototype, "makeSafeRef", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("MakeSafeRef");
        },
        enumerable: false,
        configurable: true
    });
    FinancialCalculationSupportingCalculationsDescription.$t = markType(FinancialCalculationSupportingCalculationsDescription, 'FinancialCalculationSupportingCalculationsDescription', Description.$);
    return FinancialCalculationSupportingCalculationsDescription;
}(Description));
export { FinancialCalculationSupportingCalculationsDescription };
