// igniteui-react-core/esm5/lib/FinancialPriceSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { FinancialSeriesDescription } from "./FinancialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var FinancialPriceSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialPriceSeriesDescription, _super);
    function FinancialPriceSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.go = null;
        _this.gn = null;
        _this.gb = null;
        _this.f7 = 0;
        _this.f5 = null;
        _this.f9 = 0;
        _this.gf = null;
        _this.gc = null;
        _this.gk = null;
        _this.gi = null;
        _this.gg = null;
        _this.gd = null;
        _this.gl = null;
        _this.gj = null;
        _this.gh = null;
        _this.ge = null;
        _this.gm = null;
        return _this;
    }
    FinancialPriceSeriesDescription.prototype.get_type = function () {
        return "FinancialPriceSeries";
    };
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "trendLineType", {
        get: function () {
            return this.go;
        },
        set: function (a) {
            this.go = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "displayType", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("DisplayType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "categoryCollisionMode", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("CategoryCollisionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "negativeOutline", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("NegativeOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "lowMemberAsLegendLabel", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("LowMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "highMemberAsLegendLabel", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("HighMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "closeMemberAsLegendLabel", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("CloseMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "openMemberAsLegendLabel", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("OpenMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "lowMemberAsLegendUnit", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("LowMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "highMemberAsLegendUnit", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("HighMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "closeMemberAsLegendUnit", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("CloseMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialPriceSeriesDescription.prototype, "openMemberAsLegendUnit", {
        get: function () {
            return this.gm;
        },
        set: function (a) {
            this.gm = a;
            this.g("OpenMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    FinancialPriceSeriesDescription.$t = markType(FinancialPriceSeriesDescription, 'FinancialPriceSeriesDescription', FinancialSeriesDescription.$);
    return FinancialPriceSeriesDescription;
}(FinancialSeriesDescription));
export { FinancialPriceSeriesDescription };
