// igniteui-react-core/esm5/lib/FinancialIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { FinancialSeriesDescription } from "./FinancialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var FinancialIndicatorDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialIndicatorDescription, _super);
    function FinancialIndicatorDescription() {
        var _this = _super.call(this) || this;
        _this.ge = null;
        _this.f9 = 0;
        _this.gg = null;
        _this.gf = null;
        _this.gd = null;
        _this.f7 = 0;
        _this.f5 = null;
        _this.ga = 0;
        return _this;
    }
    FinancialIndicatorDescription.prototype.get_type = function () {
        return "FinancialIndicator";
    };
    Object.defineProperty(FinancialIndicatorDescription.prototype, "displayType", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("DisplayType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "ignoreFirst", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("IgnoreFirst");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "trendLineType", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialIndicatorDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    FinancialIndicatorDescription.$t = markType(FinancialIndicatorDescription, 'FinancialIndicatorDescription', FinancialSeriesDescription.$);
    return FinancialIndicatorDescription;
}(FinancialSeriesDescription));
export { FinancialIndicatorDescription };
