// igniteui-react-core/esm5/lib/FlatDataProviderJoinCollisionType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Enumerates the join types available to the flat data provider.
 */
export var FlatDataProviderJoinCollisionType = /*@__PURE__*/ (function (FlatDataProviderJoinCollisionType) {
    /**
     * Retention is based on the join type.
     */
    FlatDataProviderJoinCollisionType[FlatDataProviderJoinCollisionType["Auto"] = 0] = "Auto";
    /**
     * Retains values only if they are from left.
     */
    FlatDataProviderJoinCollisionType[FlatDataProviderJoinCollisionType["LeftOnly"] = 1] = "LeftOnly";
    /**
     * Retains values only if they are from right.
     */
    FlatDataProviderJoinCollisionType[FlatDataProviderJoinCollisionType["RightOnly"] = 2] = "RightOnly";
    /**
     * Prefers to retain values from left.
     */
    FlatDataProviderJoinCollisionType[FlatDataProviderJoinCollisionType["PreferLeft"] = 3] = "PreferLeft";
    /**
     * Prefers to retain values from right.
     */
    FlatDataProviderJoinCollisionType[FlatDataProviderJoinCollisionType["PreferRight"] = 4] = "PreferRight";
    return FlatDataProviderJoinCollisionType;
})({});
/**
 * @hidden
 */
export var FlatDataProviderJoinCollisionType_$type = /*@__PURE__*/ markEnum('FlatDataProviderJoinCollisionType', 'Auto,0|LeftOnly,1|RightOnly,2|PreferLeft,3|PreferRight,4');
