// igniteui-react-core/esm5/lib/FormatSummaryTextEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var FormatSummaryTextEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FormatSummaryTextEventArgsDescription, _super);
    function FormatSummaryTextEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.i = null;
        _this.h = null;
        _this.l = null;
        _this.j = null;
        _this.k = null;
        return _this;
    }
    FormatSummaryTextEventArgsDescription.prototype.get_type = function () {
        return "FormatSummaryTextEventArgs";
    };
    Object.defineProperty(FormatSummaryTextEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FormatSummaryTextEventArgsDescription.prototype, "field", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Field");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FormatSummaryTextEventArgsDescription.prototype, "displayName", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("DisplayName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FormatSummaryTextEventArgsDescription.prototype, "summaryResultRef", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("SummaryResultRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FormatSummaryTextEventArgsDescription.prototype, "formattedResult", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("FormattedResult");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FormatSummaryTextEventArgsDescription.prototype, "formattedText", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("FormattedText");
        },
        enumerable: false,
        configurable: true
    });
    FormatSummaryTextEventArgsDescription.$t = markType(FormatSummaryTextEventArgsDescription, 'FormatSummaryTextEventArgsDescription', Description.$);
    return FormatSummaryTextEventArgsDescription;
}(Description));
export { FormatSummaryTextEventArgsDescription };
