// igniteui-react-core/esm5/lib/FormatGroupTextEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var FormatGroupTextEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FormatGroupTextEventArgsDescription, _super);
    function FormatGroupTextEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.k = null;
        _this.i = null;
        _this.h = null;
        return _this;
    }
    FormatGroupTextEventArgsDescription.prototype.get_type = function () {
        return "FormatGroupTextEventArgs";
    };
    Object.defineProperty(FormatGroupTextEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FormatGroupTextEventArgsDescription.prototype, "groupName", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("GroupName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FormatGroupTextEventArgsDescription.prototype, "groupValueRef", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("GroupValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FormatGroupTextEventArgsDescription.prototype, "formattedValue", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("FormattedValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FormatGroupTextEventArgsDescription.prototype, "formattedText", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("FormattedText");
        },
        enumerable: false,
        configurable: true
    });
    FormatGroupTextEventArgsDescription.$t = markType(FormatGroupTextEventArgsDescription, 'FormatGroupTextEventArgsDescription', Description.$);
    return FormatGroupTextEventArgsDescription;
}(Description));
export { FormatGroupTextEventArgsDescription };
