// igniteui-react-core/esm5/lib/FunnelSliceEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var FunnelSliceEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FunnelSliceEventArgsDescription, _super);
    function FunnelSliceEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.l = 0;
        _this.j = null;
        _this.n = null;
        _this.h = null;
        return _this;
    }
    FunnelSliceEventArgsDescription.prototype.get_type = function () {
        return "FunnelSliceEventArgs";
    };
    Object.defineProperty(FunnelSliceEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FunnelSliceEventArgsDescription.prototype, "index", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Index");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FunnelSliceEventArgsDescription.prototype, "bounds", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Bounds");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FunnelSliceEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FunnelSliceEventArgsDescription.prototype, "position", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Position");
        },
        enumerable: false,
        configurable: true
    });
    FunnelSliceEventArgsDescription.$t = markType(FunnelSliceEventArgsDescription, 'FunnelSliceEventArgsDescription', Description.$);
    return FunnelSliceEventArgsDescription;
}(Description));
export { FunnelSliceEventArgsDescription };
