// igniteui-react-core/esm5/lib/GeographicHighDensityScatterSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicHighDensityScatterSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicHighDensityScatterSeriesDescription, _super);
    function GeographicHighDensityScatterSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fl = null;
        _this.fm = null;
        _this.e8 = false;
        _this.e7 = false;
        _this.fc = 0;
        _this.fb = 0;
        _this.fk = null;
        _this.fj = null;
        _this.ff = 0;
        _this.fg = 0;
        _this.fn = null;
        return _this;
    }
    GeographicHighDensityScatterSeriesDescription.prototype.get_type = function () {
        return "GeographicHighDensityScatterSeries";
    };
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "latitudeMemberPath", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("LatitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "longitudeMemberPath", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("LongitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "useBruteForce", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("UseBruteForce");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "progressiveLoad", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("ProgressiveLoad");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMinimum", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("HeatMinimum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMaximum", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("HeatMaximum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMinimumColor", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("HeatMinimumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMaximumColor", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("HeatMaximumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "pointExtent", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            this.ff = a;
            this.g("PointExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "progressiveStatus", {
        get: function () {
            return this.fg;
        },
        set: function (a) {
            this.fg = a;
            this.g("ProgressiveStatus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "progressiveLoadStatusChangedRef", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("ProgressiveLoadStatusChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicHighDensityScatterSeriesDescription.$t = markType(GeographicHighDensityScatterSeriesDescription, 'GeographicHighDensityScatterSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicHighDensityScatterSeriesDescription;
}(GeographicMapSeriesHostDescription));
export { GeographicHighDensityScatterSeriesDescription };
