// igniteui-react-core/esm5/lib/GeographicXYTriangulatingSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicXYTriangulatingSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicXYTriangulatingSeriesDescription, _super);
    function GeographicXYTriangulatingSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.e8 = null;
        _this.e7 = null;
        _this.e9 = null;
        _this.fa = null;
        _this.fb = null;
        _this.fc = null;
        return _this;
    }
    GeographicXYTriangulatingSeriesDescription.prototype.get_type = function () {
        return "GeographicXYTriangulatingSeries";
    };
    Object.defineProperty(GeographicXYTriangulatingSeriesDescription.prototype, "longitudeMemberPath", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("LongitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicXYTriangulatingSeriesDescription.prototype, "latitudeMemberPath", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("LatitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicXYTriangulatingSeriesDescription.prototype, "trianglesSourceRef", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("TrianglesSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicXYTriangulatingSeriesDescription.prototype, "triangleVertexMemberPath1", {
        get: function () {
            return this.fa;
        },
        set: function (a) {
            this.fa = a;
            this.g("TriangleVertexMemberPath1");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicXYTriangulatingSeriesDescription.prototype, "triangleVertexMemberPath2", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("TriangleVertexMemberPath2");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicXYTriangulatingSeriesDescription.prototype, "triangleVertexMemberPath3", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("TriangleVertexMemberPath3");
        },
        enumerable: false,
        configurable: true
    });
    GeographicXYTriangulatingSeriesDescription.$t = markType(GeographicXYTriangulatingSeriesDescription, 'GeographicXYTriangulatingSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicXYTriangulatingSeriesDescription;
}(GeographicMapSeriesHostDescription));
export { GeographicXYTriangulatingSeriesDescription };
