// igniteui-react-core/esm5/lib/GeographicShapeSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicShapeSeriesBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicShapeSeriesBaseDescription, _super);
    function GeographicShapeSeriesBaseDescription() {
        var _this = _super.call(this) || this;
        _this.fr = null;
        _this.fd = false;
        _this.fc = false;
        _this.fq = null;
        _this.fk = 0;
        _this.fp = null;
        _this.e7 = null;
        _this.fb = false;
        _this.fj = 0;
        _this.e9 = null;
        _this.fh = 0;
        _this.fo = null;
        _this.fn = null;
        return _this;
    }
    GeographicShapeSeriesBaseDescription.prototype.get_type = function () {
        return "GeographicShapeSeriesBase";
    };
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "shapeMemberPath", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("ShapeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "isCustomShapeStyleAllowed", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("IsCustomShapeStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "isCustomShapeMarkerStyleAllowed", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("IsCustomShapeMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "itemSearchMode", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("ItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "itemSearchThreshold", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("ItemSearchThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "fillMemberPath", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("FillMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "fillScale", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("FillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "fillScaleUseGlobalValues", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("FillScaleUseGlobalValues");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "itemSearchPointsThreshold", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("ItemSearchPointsThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "shapefileDataSource", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("ShapefileDataSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "shapeFilterResolution", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("ShapeFilterResolution");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "assigningShapeStyleRef", {
        get: function () {
            return this.fo;
        },
        set: function (a) {
            this.fo = a;
            this.g("AssigningShapeStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "assigningShapeMarkerStyleRef", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("AssigningShapeMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicShapeSeriesBaseDescription.$t = markType(GeographicShapeSeriesBaseDescription, 'GeographicShapeSeriesBaseDescription', GeographicMapSeriesHostDescription.$);
    return GeographicShapeSeriesBaseDescription;
}(GeographicMapSeriesHostDescription));
export { GeographicShapeSeriesBaseDescription };
