// igniteui-react-core/esm5/lib/GeographicPolylineSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicShapeSeriesBaseDescription } from "./GeographicShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicPolylineSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicPolylineSeriesDescription, _super);
    function GeographicPolylineSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.f3 = null;
        _this.f1 = null;
        _this.f2 = null;
        _this.fy = 0;
        _this.fx = 0;
        _this.f4 = null;
        return _this;
    }
    GeographicPolylineSeriesDescription.prototype.get_type = function () {
        return "GeographicPolylineSeries";
    };
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeStyleSelectorRef", {
        get: function () {
            return this.f3;
        },
        set: function (a) {
            this.f3 = a;
            this.g("ShapeStyleSelectorRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeFill", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("ShapeFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeStroke", {
        get: function () {
            return this.f2;
        },
        set: function (a) {
            this.f2 = a;
            this.g("ShapeStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeStrokeThickness", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            this.fy = a;
            this.g("ShapeStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeOpacity", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("ShapeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "styleShapeRef", {
        get: function () {
            return this.f4;
        },
        set: function (a) {
            this.f4 = a;
            this.g("StyleShapeRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicPolylineSeriesDescription.$t = markType(GeographicPolylineSeriesDescription, 'GeographicPolylineSeriesDescription', GeographicShapeSeriesBaseDescription.$);
    return GeographicPolylineSeriesDescription;
}(GeographicShapeSeriesBaseDescription));
export { GeographicPolylineSeriesDescription };
