// igniteui-react-core/esm5/lib/GlobalAnimationState.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, runOn, markType } from "./type";
import { VersionQueueHelper } from "./VersionQueueHelper";
/**
 * @hidden
 */
var GlobalAnimationState = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GlobalAnimationState, _super);
    function GlobalAnimationState() {
        var _this = _super.call(this) || this;
        _this.h = 0;
        _this.a = null;
        _this.f = false;
        _this.e = new VersionQueueHelper();
        return _this;
    }
    Object.defineProperty(GlobalAnimationState, "d", {
        get: function () {
            if (GlobalAnimationState.c == null) {
                GlobalAnimationState.c = new GlobalAnimationState();
            }
            return GlobalAnimationState.c;
        },
        enumerable: false,
        configurable: true
    });
    GlobalAnimationState.prototype.p = function (a) {
        this.a = a;
        this.e.l(a);
    };
    GlobalAnimationState.prototype.b = function () {
        if (this.a != null) {
            return this.a;
        }
        return null;
    };
    GlobalAnimationState.prototype.l = function () {
        this.h++;
    };
    GlobalAnimationState.prototype.k = function () {
        if (this.h > 0) {
            this.h--;
        }
        if (this.h <= 0) {
            this.h = 0;
            this.o();
        }
    };
    GlobalAnimationState.prototype.o = function () {
        if (this.f) {
            return;
        }
        var a = this.b();
        if (a == null) {
            this.e.h();
            return;
        }
        a.executeDelayed(runOn(this, this.j), 0);
        this.f = true;
    };
    GlobalAnimationState.prototype.j = function () {
        if (this.h <= 0) {
            this.h = 0;
            this.e.h();
        }
        this.f = false;
    };
    GlobalAnimationState.prototype.m = function (a, b) {
        this.e.j(a, b);
    };
    GlobalAnimationState.prototype.n = function (a, b, c) {
        this.e.k(a, b, c);
    };
    GlobalAnimationState.prototype.i = function () {
        return this.e.f();
    };
    GlobalAnimationState.prototype.g = function () {
        return this.h > 0;
    };
    GlobalAnimationState.$t = markType(GlobalAnimationState, 'GlobalAnimationState');
    GlobalAnimationState.c = null;
    return GlobalAnimationState;
}(Base));
export { GlobalAnimationState };
