// igniteui-react-core/esm5/lib/GridColumnOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GridColumnOptionsBaseDescription } from "./GridColumnOptionsBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GridColumnOptionsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GridColumnOptionsDescription, _super);
    function GridColumnOptionsDescription() {
        var _this = _super.call(this) || this;
        _this.bp = null;
        _this.a8 = 0;
        _this.bq = null;
        _this.br = null;
        _this.bb = 0;
        _this.bd = 0;
        _this.bc = 0;
        _this.ba = 0;
        _this.au = false;
        _this.ay = false;
        _this.aw = false;
        _this.ax = false;
        _this.av = false;
        _this.at = false;
        _this.as = false;
        _this.az = false;
        _this.b2 = null;
        _this.b0 = null;
        _this.b1 = null;
        _this.bu = null;
        _this.bv = null;
        _this.bw = null;
        _this.bx = null;
        _this.by = null;
        _this.bz = null;
        _this.bl = null;
        _this.bn = null;
        _this.bo = null;
        _this.b4 = null;
        _this.bj = null;
        _this.bs = null;
        _this.bi = null;
        _this.bt = null;
        _this.b5 = null;
        _this.b3 = null;
        _this.bk = null;
        _this.bm = null;
        return _this;
    }
    GridColumnOptionsDescription.prototype.get_type = function () {
        return "GridColumnOptions";
    };
    Object.defineProperty(GridColumnOptionsDescription.prototype, "columnNameFontFamily", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.g("ColumnNameFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "columnNameFontSize", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("ColumnNameFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "columnNameFontStyle", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            this.bq = a;
            this.g("ColumnNameFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "columnNameFontWeight", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.g("ColumnNameFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "paddingLeft", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("PaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "paddingTop", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("PaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "paddingRight", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("PaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "paddingBottom", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("PaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "headerVisible", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("HeaderVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "sortOptionsVisible", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("SortOptionsVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "moveOptionsVisible", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("MoveOptionsVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "pinOptionsVisible", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("PinOptionsVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "hideOptionsVisible", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("HideOptionsVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "groupOptionsVisible", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("GroupOptionsVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "filterOptionsVisible", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("FilterOptionsVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "summaryOptionsVisible", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("SummaryOptionsVisible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "sortHeaderCaption", {
        get: function () {
            return this.b2;
        },
        set: function (a) {
            this.b2 = a;
            this.g("SortHeaderCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "sortAscendingCaption", {
        get: function () {
            return this.b0;
        },
        set: function (a) {
            this.b0 = a;
            this.g("SortAscendingCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "sortDescendingCaption", {
        get: function () {
            return this.b1;
        },
        set: function (a) {
            this.b1 = a;
            this.g("SortDescendingCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "moveHeaderCaption", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("MoveHeaderCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "moveLeftCaption", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("MoveLeftCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "moveRightCaption", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("MoveRightCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "pinHeaderCaption", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("PinHeaderCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "pinLeftCaption", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            this.by = a;
            this.g("PinLeftCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "pinRightCaption", {
        get: function () {
            return this.bz;
        },
        set: function (a) {
            this.bz = a;
            this.g("PinRightCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "applyFiltersButtonCaption", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.g("ApplyFiltersButtonCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "cancelFiltersButtonCaption", {
        get: function () {
            return this.bn;
        },
        set: function (a) {
            this.bn = a;
            this.g("CancelFiltersButtonCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "clearColumnFiltersCaption", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            this.bo = a;
            this.g("ClearColumnFiltersCaption");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "summaryListDensity", {
        get: function () {
            return this.b4;
        },
        set: function (a) {
            this.b4 = a;
            this.g("SummaryListDensity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "actualSummaryListDensity", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("ActualSummaryListDensity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "filterListDensity", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("FilterListDensity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "actualFilterListDensity", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("ActualFilterListDensity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "filterListPlaceholderText", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("FilterListPlaceholderText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "summaryListTextColor", {
        get: function () {
            return this.b5;
        },
        set: function (a) {
            this.b5 = a;
            this.g("SummaryListTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "summaryListBackground", {
        get: function () {
            return this.b3;
        },
        set: function (a) {
            this.b3 = a;
            this.g("SummaryListBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "applyButtonClickRef", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("ApplyButtonClickRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridColumnOptionsDescription.prototype, "cancelButtonClickRef", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            this.bm = a;
            this.g("CancelButtonClickRef");
        },
        enumerable: false,
        configurable: true
    });
    GridColumnOptionsDescription.$t = markType(GridColumnOptionsDescription, 'GridColumnOptionsDescription', GridColumnOptionsBaseDescription.$);
    return GridColumnOptionsDescription;
}(GridColumnOptionsBaseDescription));
export { GridColumnOptionsDescription };
