// igniteui-react-core/esm5/lib/GridCustomFilterRequestedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var GridCustomFilterRequestedEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GridCustomFilterRequestedEventArgsDescription, _super);
    function GridCustomFilterRequestedEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.l = null;
        _this.n = null;
        _this.j = null;
        return _this;
    }
    GridCustomFilterRequestedEventArgsDescription.prototype.get_type = function () {
        return "GridCustomFilterRequestedEventArgs";
    };
    Object.defineProperty(GridCustomFilterRequestedEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCustomFilterRequestedEventArgsDescription.prototype, "column", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Column");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCustomFilterRequestedEventArgsDescription.prototype, "filter", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCustomFilterRequestedEventArgsDescription.prototype, "valueRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCustomFilterRequestedEventArgsDescription.prototype, "expression", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Expression");
        },
        enumerable: false,
        configurable: true
    });
    GridCustomFilterRequestedEventArgsDescription.$t = markType(GridCustomFilterRequestedEventArgsDescription, 'GridCustomFilterRequestedEventArgsDescription', Description.$);
    return GridCustomFilterRequestedEventArgsDescription;
}(Description));
export { GridCustomFilterRequestedEventArgsDescription };
