// igniteui-react-core/esm5/lib/GridFilterDialogViewModelRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var GridFilterDialogViewModelRowDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GridFilterDialogViewModelRowDescription, _super);
    function GridFilterDialogViewModelRowDescription() {
        var _this = _super.call(this) || this;
        _this.i = null;
        _this.l = null;
        _this.h = null;
        _this.m = null;
        _this.n = null;
        return _this;
    }
    GridFilterDialogViewModelRowDescription.prototype.get_type = function () {
        return "GridFilterDialogViewModelRow";
    };
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "operators", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Operators");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "currentOperator", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("CurrentOperator");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "operandNumber", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("OperandNumber");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "operand1Ref", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Operand1Ref");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "operand2Ref", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Operand2Ref");
        },
        enumerable: false,
        configurable: true
    });
    GridFilterDialogViewModelRowDescription.$t = markType(GridFilterDialogViewModelRowDescription, 'GridFilterDialogViewModelRowDescription', Description.$);
    return GridFilterDialogViewModelRowDescription;
}(Description));
export { GridFilterDialogViewModelRowDescription };
