// igniteui-react-core/esm5/lib/GridSelectedCellRangesChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var GridSelectedCellRangesChangedEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GridSelectedCellRangesChangedEventArgsDescription, _super);
    function GridSelectedCellRangesChangedEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.i = null;
        _this.h = null;
        _this.j = null;
        _this.k = null;
        return _this;
    }
    GridSelectedCellRangesChangedEventArgsDescription.prototype.get_type = function () {
        return "GridSelectedCellRangesChangedEventArgs";
    };
    Object.defineProperty(GridSelectedCellRangesChangedEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridSelectedCellRangesChangedEventArgsDescription.prototype, "currentRanges", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("CurrentRanges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridSelectedCellRangesChangedEventArgsDescription.prototype, "addedRanges", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("AddedRanges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridSelectedCellRangesChangedEventArgsDescription.prototype, "removedRanges", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("RemovedRanges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridSelectedCellRangesChangedEventArgsDescription.prototype, "updatedRanges", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("UpdatedRanges");
        },
        enumerable: false,
        configurable: true
    });
    GridSelectedCellRangesChangedEventArgsDescription.$t = markType(GridSelectedCellRangesChangedEventArgsDescription, 'GridSelectedCellRangesChangedEventArgsDescription', Description.$);
    return GridSelectedCellRangesChangedEventArgsDescription;
}(Description));
export { GridSelectedCellRangesChangedEventArgsDescription };
