// igniteui-react-core/esm5/lib/GridCellEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var GridCellEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GridCellEventArgsDescription, _super);
    function GridCellEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.j = null;
        _this.n = null;
        _this.l = false;
        return _this;
    }
    GridCellEventArgsDescription.prototype.get_type = function () {
        return "GridCellEventArgs";
    };
    Object.defineProperty(GridCellEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCellEventArgsDescription.prototype, "cellInfo", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("CellInfo");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCellEventArgsDescription.prototype, "grid", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Grid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCellEventArgsDescription.prototype, "button", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Button");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCellEventArgsDescription.prototype, "isDoubleClick", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IsDoubleClick");
        },
        enumerable: false,
        configurable: true
    });
    GridCellEventArgsDescription.$t = markType(GridCellEventArgsDescription, 'GridCellEventArgsDescription', Description.$);
    return GridCellEventArgsDescription;
}(Description));
export { GridCellEventArgsDescription };
