// igniteui-react-core/esm5/lib/GridCellValueChangingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var GridCellValueChangingEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GridCellValueChangingEventArgsDescription, _super);
    function GridCellValueChangingEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.l = 0;
        _this.h = null;
        _this.j = null;
        _this.r = null;
        _this.o = null;
        _this.n = null;
        return _this;
    }
    GridCellValueChangingEventArgsDescription.prototype.get_type = function () {
        return "GridCellValueChangingEventArgs";
    };
    Object.defineProperty(GridCellValueChangingEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangingEventArgsDescription.prototype, "editID", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("EditID");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangingEventArgsDescription.prototype, "cellInfo", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("CellInfo");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangingEventArgsDescription.prototype, "column", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Column");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangingEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangingEventArgsDescription.prototype, "oldValue", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("OldValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangingEventArgsDescription.prototype, "newValue", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("NewValue");
        },
        enumerable: false,
        configurable: true
    });
    GridCellValueChangingEventArgsDescription.$t = markType(GridCellValueChangingEventArgsDescription, 'GridCellValueChangingEventArgsDescription', Description.$);
    return GridCellValueChangingEventArgsDescription;
}(Description));
export { GridCellValueChangingEventArgsDescription };
