// igniteui-react-core/esm5/lib/HeaderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var HeaderDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(HeaderDescription, _super);
    function HeaderDescription() {
        var _this = _super.call(this) || this;
        _this.c0 = null;
        _this.cy = null;
        _this.c1 = null;
        _this.cz = null;
        _this.cq = 0;
        _this.cs = 0;
        _this.cr = 0;
        _this.cp = 0;
        return _this;
    }
    HeaderDescription.prototype.get_type = function () {
        return "Header";
    };
    Object.defineProperty(HeaderDescription.prototype, "sortIndicatorColor", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("SortIndicatorColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HeaderDescription.prototype, "actualSortIndicatorColor", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("ActualSortIndicatorColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HeaderDescription.prototype, "sortIndicatorStyle", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.g("SortIndicatorStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HeaderDescription.prototype, "actualSortIndicatorStyle", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("ActualSortIndicatorStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HeaderDescription.prototype, "paddingLeft", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("PaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HeaderDescription.prototype, "paddingTop", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("PaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HeaderDescription.prototype, "paddingRight", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("PaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HeaderDescription.prototype, "paddingBottom", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("PaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    HeaderDescription.$t = markType(HeaderDescription, 'HeaderDescription', DefinitionBaseDescription.$);
    HeaderDescription.__canMarshalByValue = true;
    HeaderDescription.__marshalByValueAlias = "Header";
    return HeaderDescription;
}(DefinitionBaseDescription));
export { HeaderDescription };
