// igniteui-react-core/esm5/lib/HighDensityScatterSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var HighDensityScatterSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(HighDensityScatterSeriesDescription, _super);
    function HighDensityScatterSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fi = null;
        _this.fm = null;
        _this.fl = null;
        _this.fj = null;
        _this.fn = null;
        _this.fk = null;
        _this.fo = null;
        _this.fp = null;
        _this.e4 = false;
        _this.e3 = false;
        _this.e8 = 0;
        _this.e7 = 0;
        _this.fg = null;
        _this.ff = null;
        _this.fb = 0;
        _this.fc = 0;
        _this.fh = null;
        return _this;
    }
    HighDensityScatterSeriesDescription.prototype.get_type = function () {
        return "HighDensityScatterSeries";
    };
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xAxisRef", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("XAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yAxisRef", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("YAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xMemberPath", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("XMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xMemberAsLegendLabel", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("XMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yMemberAsLegendLabel", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("YMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xMemberAsLegendUnit", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("XMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yMemberAsLegendUnit", {
        get: function () {
            return this.fo;
        },
        set: function (a) {
            this.fo = a;
            this.g("YMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yMemberPath", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("YMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "useBruteForce", {
        get: function () {
            return this.e4;
        },
        set: function (a) {
            this.e4 = a;
            this.g("UseBruteForce");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "progressiveLoad", {
        get: function () {
            return this.e3;
        },
        set: function (a) {
            this.e3 = a;
            this.g("ProgressiveLoad");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMinimum", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("HeatMinimum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMaximum", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("HeatMaximum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMinimumColor", {
        get: function () {
            return this.fg;
        },
        set: function (a) {
            this.fg = a;
            this.g("HeatMinimumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMaximumColor", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            this.ff = a;
            this.g("HeatMaximumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "pointExtent", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("PointExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "progressiveStatus", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("ProgressiveStatus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "progressiveLoadStatusChangedRef", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("ProgressiveLoadStatusChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    HighDensityScatterSeriesDescription.$t = markType(HighDensityScatterSeriesDescription, 'HighDensityScatterSeriesDescription', SeriesDescription.$);
    return HighDensityScatterSeriesDescription;
}(SeriesDescription));
export { HighDensityScatterSeriesDescription };
