// igniteui-react-core/esm5/lib/HorizontalRangeCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { RangeCategorySeriesDescription } from "./RangeCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var HorizontalRangeCategorySeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(HorizontalRangeCategorySeriesDescription, _super);
    function HorizontalRangeCategorySeriesDescription() {
        var _this = _super.call(this) || this;
        _this.gj = null;
        _this.gk = null;
        _this.gh = null;
        _this.gf = null;
        _this.gi = null;
        _this.gg = null;
        return _this;
    }
    HorizontalRangeCategorySeriesDescription.prototype.get_type = function () {
        return "HorizontalRangeCategorySeries";
    };
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "xAxisRef", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("XAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "yAxisRef", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("YAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "lowMemberAsLegendLabel", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("LowMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "highMemberAsLegendLabel", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("HighMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "lowMemberAsLegendUnit", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("LowMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "highMemberAsLegendUnit", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("HighMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    HorizontalRangeCategorySeriesDescription.$t = markType(HorizontalRangeCategorySeriesDescription, 'HorizontalRangeCategorySeriesDescription', RangeCategorySeriesDescription.$);
    return HorizontalRangeCategorySeriesDescription;
}(RangeCategorySeriesDescription));
export { HorizontalRangeCategorySeriesDescription };
