// igniteui-react-core/esm5/lib/IgcCheckboxBridge.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { NativeUIElementBridge } from "./NativeUIElementBridge";
import { runOn, markType } from "./type";
import { NativeUICheckboxChangeEventArgs } from "./NativeUICheckboxChangeEventArgs";
/**
 * @hidden
 */
var IgcCheckboxBridge = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgcCheckboxBridge, _super);
    function IgcCheckboxBridge() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.t = null;
        _this.s = null;
        _this.u = null;
        return _this;
    }
    IgcCheckboxBridge.prototype.addHandler = function (a, b, c, d) {
        switch (c) {
            case 2:
                var e = a;
                this.t = e.listen("igcChange", runOn(this, this.v));
                this.s = d;
                this.u = b;
                break;
        }
    };
    IgcCheckboxBridge.prototype.v = function (a) {
        if (this.s != null) {
            var b = new NativeUICheckboxChangeEventArgs();
            this.s(this.u, b);
        }
    };
    IgcCheckboxBridge.prototype.getValue = function (a, b) {
        switch (b) {
            case 4: return true;
            case 7: return a.getProperty("checked");
        }
        return null;
    };
    IgcCheckboxBridge.prototype.removeHandler = function (a, b, c, d) {
        switch (c) {
            case 2:
                if (this.t != null) {
                    this.t();
                }
                this.s = null;
                this.u = null;
                break;
        }
    };
    IgcCheckboxBridge.prototype.setValue = function (a, b, c) {
        switch (b) {
            case 4: return;
            case 7:
                a.setProperty("checked", c);
                return;
        }
    };
    IgcCheckboxBridge.$t = markType(IgcCheckboxBridge, 'IgcCheckboxBridge', NativeUIElementBridge.$);
    return IgcCheckboxBridge;
}(NativeUIElementBridge));
export { IgcCheckboxBridge };
