// igniteui-react-core/esm5/lib/igr-format-specifier.js
import { FormatSpecifier as FormatSpecifier_internal } from "./FormatSpecifier";
var IgrFormatSpecifier = /** @class */ /*@__PURE__*/ (function () {
    function IgrFormatSpecifier() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrFormatSpecifier.prototype.createImplementation = function () {
        return new FormatSpecifier_internal();
    };
    Object.defineProperty(IgrFormatSpecifier.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrFormatSpecifier.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrFormatSpecifier.prototype.onImplementationCreated = function () {
    };
    IgrFormatSpecifier.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    IgrFormatSpecifier.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    IgrFormatSpecifier.prototype.getLocalCulture = function () {
        var iv = this.i.getLocalCulture();
        return (iv);
    };
    return IgrFormatSpecifier;
}());
export { IgrFormatSpecifier };
