// igniteui-react-core/esm5/lib/igr-data-legend-series-value-info.js
import { DataLegendSeriesValueType_$type } from "./DataLegendSeriesValueType";
import { DataLegendSeriesValueInfo as DataLegendSeriesValueInfo_internal } from "./DataLegendSeriesValueInfo";
import { brushToString, stringToBrush, ensureBool, ensureEnum } from "./componentUtil";
/**
 * Represent info about series values
*/
var IgrDataLegendSeriesValueInfo = /** @class */ /*@__PURE__*/ (function () {
    function IgrDataLegendSeriesValueInfo() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrDataLegendSeriesValueInfo.prototype.createImplementation = function () {
        return new DataLegendSeriesValueInfo_internal();
    };
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrDataLegendSeriesValueInfo.prototype.onImplementationCreated = function () {
    };
    IgrDataLegendSeriesValueInfo.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "value", {
        /**
         * Gets value of series
        */
        get: function () {
            return this.i.l;
        },
        set: function (v) {
            this.i.l = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "valueNegativePrefix", {
        /**
         * Gets or sets prefix for negative value
        */
        get: function () {
            return this.i.v;
        },
        set: function (v) {
            this.i.v = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "valueNegativeSuffix", {
        /**
         * Gets or sets suffix for negative value
        */
        get: function () {
            return this.i.w;
        },
        set: function (v) {
            this.i.w = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "valuePositivePrefix", {
        /**
         * Gets or sets prefix for positive value
        */
        get: function () {
            return this.i.x;
        },
        set: function (v) {
            this.i.x = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "valuePositiveSuffix", {
        /**
         * Gets or sets suffix for positive value
        */
        get: function () {
            return this.i.y;
        },
        set: function (v) {
            this.i.y = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "formatWithSeriesColor", {
        /**
         * Gets value color used by the series
        */
        get: function () {
            return brushToString(this.i.z);
        },
        set: function (v) {
            this.i.z = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "formatMinFractions", {
        /**
         * Gets or sets minimum digits used for displaying faction of the value
        */
        get: function () {
            return this.i.n;
        },
        set: function (v) {
            this.i.n = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "formatMaxFractions", {
        /**
         * Gets or sets maximum digits used for displaying faction of the value
        */
        get: function () {
            return this.i.m;
        },
        set: function (v) {
            this.i.m = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "formatUsePositiveColor", {
        /**
         * Gets or sets whether or not use conditional color
        */
        get: function () {
            return this.i.j;
        },
        set: function (v) {
            this.i.j = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "formatUseNegativeColor", {
        /**
         * Gets or sets whether or not use conditional color
        */
        get: function () {
            return this.i.i;
        },
        set: function (v) {
            this.i.i = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "formatAllowCurrency", {
        /**
         * Gets or sets whether or not allow displaying value as currency
        */
        get: function () {
            return this.i.e;
        },
        set: function (v) {
            this.i.e = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "formatAllowPercent", {
        /**
         * Gets or sets whether or not allow displaying value as percentage
        */
        get: function () {
            return this.i.h;
        },
        set: function (v) {
            this.i.h = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "formatAllowDecimal", {
        /**
         * Gets or sets whether or not allow displaying value as decimal
        */
        get: function () {
            return this.i.f;
        },
        set: function (v) {
            this.i.f = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "formatAllowInteger", {
        /**
         * Gets or sets whether or not allow displaying value as integer
        */
        get: function () {
            return this.i.g;
        },
        set: function (v) {
            this.i.g = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "formatAllowAbbreviation", {
        /**
         * Gets or sets whether or not allow displaying value with abbreviation
        */
        get: function () {
            return this.i.d;
        },
        set: function (v) {
            this.i.d = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "valueType", {
        /**
         * Gets or sets type of value
        */
        get: function () {
            return this.i.a;
        },
        set: function (v) {
            this.i.a = ensureEnum(DataLegendSeriesValueType_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "index", {
        /**
         * Gets the index of series
        */
        get: function () {
            return this.i.o;
        },
        set: function (v) {
            this.i.o = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "orderIndex", {
        get: function () {
            return this.i.p;
        },
        set: function (v) {
            this.i.p = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "isExcludeByDefault", {
        /**
         * Gets or sets whether this value is excluded/hidden by default in Data Legend
        */
        get: function () {
            return this.i.k;
        },
        set: function (v) {
            this.i.k = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "allowUnits", {
        /**
         * Gets or sets whether the unit text is allowed by the series, e.g. "Hz"
        */
        get: function () {
            return this.i.c;
        },
        set: function (v) {
            this.i.c = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "allowLabels", {
        /**
         * Gets or sets whether the label text is allowed by the series, e.g. "H:"
        */
        get: function () {
            return this.i.b;
        },
        set: function (v) {
            this.i.b = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "memberPath", {
        /**
         * Gets or sets member path for the series value, e.g. "HighPrice"
        */
        get: function () {
            return this.i.r;
        },
        set: function (v) {
            this.i.r = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "memberLabel", {
        /**
         * Gets or sets member path for the series value, e.g. "High"
        */
        get: function () {
            return this.i.q;
        },
        set: function (v) {
            this.i.q = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "memberSymbol", {
        /**
         * Gets or sets label for the series value, e.g. "H:"
        */
        get: function () {
            return this.i.s;
        },
        set: function (v) {
            this.i.s = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDataLegendSeriesValueInfo.prototype, "memberUnit", {
        /**
         * Gets or sets unit text for the value, e.g. "Hz"
        */
        get: function () {
            return this.i.t;
        },
        set: function (v) {
            this.i.t = v;
        },
        enumerable: false,
        configurable: true
    });
    IgrDataLegendSeriesValueInfo.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    /**
     * Converts this object to a string
    
    */
    IgrDataLegendSeriesValueInfo.prototype.toString = function () {
        var iv = this.i.u();
        return (iv);
    };
    return IgrDataLegendSeriesValueInfo;
}());
export { IgrDataLegendSeriesValueInfo };
