// igniteui-react-core/esm5/lib/igr-style.js
import { Style as Style_internal } from "./Style";
/**
 * A class that exposes style realted properties.
*/
var IgrStyle = /** @class */ /*@__PURE__*/ (function () {
    function IgrStyle() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrStyle.prototype.createImplementation = function () {
        return new Style_internal();
    };
    Object.defineProperty(IgrStyle.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStyle.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrStyle.prototype.onImplementationCreated = function () {
    };
    IgrStyle.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    IgrStyle.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    return IgrStyle;
}());
export { IgrStyle };
