// igniteui-react-core/esm5/lib/InputGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var InputGroupDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(InputGroupDescription, _super);
    function InputGroupDescription() {
        var _this = _super.call(this) || this;
        _this.gu = null;
        _this.gt = null;
        _this.f6 = null;
        _this.gd = null;
        _this.f5 = null;
        _this.f7 = null;
        _this.t = 0;
        _this.ac = 0;
        _this.z = 0;
        _this.y = 0;
        _this.aa = 0;
        _this.ab = 0;
        _this.u = 0;
        _this.v = 0;
        _this.w = 0;
        _this.x = 0;
        _this.ak = 0;
        _this.aj = 0;
        _this.ag = 0;
        _this.ga = null;
        _this.gb = null;
        _this.f9 = null;
        _this.f8 = null;
        _this.ah = 0;
        _this.af = 0;
        _this.ad = 0;
        _this.ae = 0;
        _this.ai = 0;
        _this.f4 = null;
        _this.a3 = 0;
        _this.bu = 0;
        _this.ge = null;
        _this.br = 0;
        _this.bq = 0;
        _this.bs = 0;
        _this.bt = 0;
        _this.bm = 0;
        _this.bn = 0;
        _this.bo = 0;
        _this.bp = 0;
        _this.c2 = 0;
        _this.c1 = 0;
        _this.by = 0;
        _this.hd = null;
        _this.he = null;
        _this.gx = null;
        _this.gw = null;
        _this.cz = 0;
        _this.bx = 0;
        _this.bv = 0;
        _this.bw = 0;
        _this.c0 = 0;
        _this.gc = null;
        _this.bz = 0;
        _this.b8 = 0;
        _this.g1 = null;
        _this.g0 = null;
        _this.b5 = 0;
        _this.b4 = 0;
        _this.b6 = 0;
        _this.b7 = 0;
        _this.b0 = 0;
        _this.b1 = 0;
        _this.b2 = 0;
        _this.b3 = 0;
        _this.cg = 0;
        _this.cf = 0;
        _this.cc = 0;
        _this.g4 = null;
        _this.g5 = null;
        _this.g3 = null;
        _this.g2 = null;
        _this.cd = 0;
        _this.cb = 0;
        _this.b9 = 0;
        _this.ca = 0;
        _this.ce = 0;
        _this.gz = null;
        _this.a4 = 0;
        _this.bd = 0;
        _this.go = null;
        _this.gn = null;
        _this.ba = 0;
        _this.a9 = 0;
        _this.bb = 0;
        _this.bc = 0;
        _this.a5 = 0;
        _this.a6 = 0;
        _this.a7 = 0;
        _this.a8 = 0;
        _this.bl = 0;
        _this.bk = 0;
        _this.bh = 0;
        _this.gr = null;
        _this.gs = null;
        _this.gq = null;
        _this.gp = null;
        _this.bi = 0;
        _this.bg = 0;
        _this.be = 0;
        _this.bf = 0;
        _this.bj = 0;
        _this.gm = null;
        _this.al = 0;
        _this.au = 0;
        _this.gh = null;
        _this.gg = null;
        _this.ar = 0;
        _this.aq = 0;
        _this.as = 0;
        _this.at = 0;
        _this.am = 0;
        _this.an = 0;
        _this.ao = 0;
        _this.ap = 0;
        _this.a2 = 0;
        _this.a1 = 0;
        _this.ay = 0;
        _this.gk = null;
        _this.gl = null;
        _this.gj = null;
        _this.gi = null;
        _this.az = 0;
        _this.ax = 0;
        _this.av = 0;
        _this.aw = 0;
        _this.a0 = 0;
        _this.gf = null;
        _this.ch = 0;
        _this.cq = 0;
        _this.g8 = null;
        _this.g7 = null;
        _this.cn = 0;
        _this.cm = 0;
        _this.co = 0;
        _this.cp = 0;
        _this.ci = 0;
        _this.cj = 0;
        _this.ck = 0;
        _this.cl = 0;
        _this.cy = 0;
        _this.cx = 0;
        _this.cu = 0;
        _this.hb = null;
        _this.hc = null;
        _this.ha = null;
        _this.g9 = null;
        _this.cv = 0;
        _this.ct = 0;
        _this.cr = 0;
        _this.cs = 0;
        _this.cw = 0;
        _this.g6 = null;
        _this.l = false;
        _this.n = false;
        _this.m = false;
        _this.j = false;
        _this.gv = null;
        _this.gy = null;
        _this.k = false;
        _this.h = null;
        return _this;
    }
    InputGroupDescription.prototype.get_type = function () {
        return "InputGroup";
    };
    Object.defineProperty(InputGroupDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "displayType", {
        get: function () {
            return this.gu;
        },
        set: function (a) {
            this.gu = a;
            this.g("DisplayType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "density", {
        get: function () {
            return this.gt;
        },
        set: function (a) {
            this.gt = a;
            this.g("Density");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualDensity", {
        get: function () {
            return this.f6;
        },
        set: function (a) {
            this.f6 = a;
            this.g("ActualDensity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "baseTheme", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualBorderColor", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("ActualBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualFocusBorderColor", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("ActualFocusBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualBorderWidth", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("ActualBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualFocusBorderWidth", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("ActualFocusBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualCornerRadiusBottomRight", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("ActualCornerRadiusBottomRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualCornerRadiusBottomLeft", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("ActualCornerRadiusBottomLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualCornerRadiusTopLeft", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("ActualCornerRadiusTopLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualCornerRadiusTopRight", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("ActualCornerRadiusTopRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualContentPaddingBottom", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("ActualContentPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualContentPaddingLeft", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("ActualContentPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualContentPaddingRight", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("ActualContentPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualContentPaddingTop", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("ActualContentPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualUnderlineWidth", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("ActualUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualUnderlineRippleWidth", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("ActualUnderlineRippleWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualHoverUnderlineWidth", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("ActualHoverUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualUnderlineColor", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("ActualUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualUnderlineRippleColor", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("ActualUnderlineRippleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualHoverUnderlineColor", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("ActualHoverUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualFocusUnderlineColor", {
        get: function () {
            return this.f8;
        },
        set: function (a) {
            this.f8 = a;
            this.g("ActualFocusUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualUnderlineOpacity", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("ActualUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualHoverUnderlineOpacity", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("ActualHoverUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualFocusUnderlineOpacity", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("ActualFocusUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualFocusUnderlineRippleOpacity", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("ActualFocusUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualUnderlineRippleOpacity", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("ActualUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualBackgroundColor", {
        get: function () {
            return this.f4;
        },
        set: function (a) {
            this.f4 = a;
            this.g("ActualBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderWidth", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("BorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "focusBorderWidth", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("FocusBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderColor", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("BorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "cornerRadiusBottomRight", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.g("CornerRadiusBottomRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "cornerRadiusBottomLeft", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            this.bq = a;
            this.g("CornerRadiusBottomLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "cornerRadiusTopLeft", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("CornerRadiusTopLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "cornerRadiusTopRight", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("CornerRadiusTopRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "contentPaddingBottom", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            this.bm = a;
            this.g("ContentPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "contentPaddingLeft", {
        get: function () {
            return this.bn;
        },
        set: function (a) {
            this.bn = a;
            this.g("ContentPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "contentPaddingRight", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            this.bo = a;
            this.g("ContentPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "contentPaddingTop", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.g("ContentPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "underlineWidth", {
        get: function () {
            return this.c2;
        },
        set: function (a) {
            this.c2 = a;
            this.g("UnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "underlineRippleWidth", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.g("UnderlineRippleWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "hoverUnderlineWidth", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            this.by = a;
            this.g("HoverUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "underlineColor", {
        get: function () {
            return this.hd;
        },
        set: function (a) {
            this.hd = a;
            this.g("UnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "underlineRippleColor", {
        get: function () {
            return this.he;
        },
        set: function (a) {
            this.he = a;
            this.g("UnderlineRippleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "hoverUnderlineColor", {
        get: function () {
            return this.gx;
        },
        set: function (a) {
            this.gx = a;
            this.g("HoverUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "focusUnderlineColor", {
        get: function () {
            return this.gw;
        },
        set: function (a) {
            this.gw = a;
            this.g("FocusUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "underlineOpacity", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("UnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "hoverUnderlineOpacity", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("HoverUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "focusUnderlineOpacity", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("FocusUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "focusUnderlineRippleOpacity", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("FocusUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "underlineRippleOpacity", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("UnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "backgroundColor", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("BackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeBorderWidth", {
        get: function () {
            return this.bz;
        },
        set: function (a) {
            this.bz = a;
            this.g("LineTypeBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeFocusBorderWidth", {
        get: function () {
            return this.b8;
        },
        set: function (a) {
            this.b8 = a;
            this.g("LineTypeFocusBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeFocusBorderColor", {
        get: function () {
            return this.g1;
        },
        set: function (a) {
            this.g1 = a;
            this.g("LineTypeFocusBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeBorderColor", {
        get: function () {
            return this.g0;
        },
        set: function (a) {
            this.g0 = a;
            this.g("LineTypeBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeCornerRadiusBottomRight", {
        get: function () {
            return this.b5;
        },
        set: function (a) {
            this.b5 = a;
            this.g("LineTypeCornerRadiusBottomRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeCornerRadiusBottomLeft", {
        get: function () {
            return this.b4;
        },
        set: function (a) {
            this.b4 = a;
            this.g("LineTypeCornerRadiusBottomLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeCornerRadiusTopLeft", {
        get: function () {
            return this.b6;
        },
        set: function (a) {
            this.b6 = a;
            this.g("LineTypeCornerRadiusTopLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeCornerRadiusTopRight", {
        get: function () {
            return this.b7;
        },
        set: function (a) {
            this.b7 = a;
            this.g("LineTypeCornerRadiusTopRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeContentPaddingBottom", {
        get: function () {
            return this.b0;
        },
        set: function (a) {
            this.b0 = a;
            this.g("LineTypeContentPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeContentPaddingLeft", {
        get: function () {
            return this.b1;
        },
        set: function (a) {
            this.b1 = a;
            this.g("LineTypeContentPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeContentPaddingRight", {
        get: function () {
            return this.b2;
        },
        set: function (a) {
            this.b2 = a;
            this.g("LineTypeContentPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeContentPaddingTop", {
        get: function () {
            return this.b3;
        },
        set: function (a) {
            this.b3 = a;
            this.g("LineTypeContentPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeUnderlineWidth", {
        get: function () {
            return this.cg;
        },
        set: function (a) {
            this.cg = a;
            this.g("LineTypeUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeUnderlineRippleWidth", {
        get: function () {
            return this.cf;
        },
        set: function (a) {
            this.cf = a;
            this.g("LineTypeUnderlineRippleWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeHoverUnderlineWidth", {
        get: function () {
            return this.cc;
        },
        set: function (a) {
            this.cc = a;
            this.g("LineTypeHoverUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeUnderlineColor", {
        get: function () {
            return this.g4;
        },
        set: function (a) {
            this.g4 = a;
            this.g("LineTypeUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeUnderlineRippleColor", {
        get: function () {
            return this.g5;
        },
        set: function (a) {
            this.g5 = a;
            this.g("LineTypeUnderlineRippleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeHoverUnderlineColor", {
        get: function () {
            return this.g3;
        },
        set: function (a) {
            this.g3 = a;
            this.g("LineTypeHoverUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeFocusUnderlineColor", {
        get: function () {
            return this.g2;
        },
        set: function (a) {
            this.g2 = a;
            this.g("LineTypeFocusUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeUnderlineOpacity", {
        get: function () {
            return this.cd;
        },
        set: function (a) {
            this.cd = a;
            this.g("LineTypeUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeHoverUnderlineOpacity", {
        get: function () {
            return this.cb;
        },
        set: function (a) {
            this.cb = a;
            this.g("LineTypeHoverUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeFocusUnderlineOpacity", {
        get: function () {
            return this.b9;
        },
        set: function (a) {
            this.b9 = a;
            this.g("LineTypeFocusUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeFocusUnderlineRippleOpacity", {
        get: function () {
            return this.ca;
        },
        set: function (a) {
            this.ca = a;
            this.g("LineTypeFocusUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeUnderlineRippleOpacity", {
        get: function () {
            return this.ce;
        },
        set: function (a) {
            this.ce = a;
            this.g("LineTypeUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "lineTypeBackgroundColor", {
        get: function () {
            return this.gz;
        },
        set: function (a) {
            this.gz = a;
            this.g("LineTypeBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeBorderWidth", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("BoxTypeBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeFocusBorderWidth", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("BoxTypeFocusBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeFocusBorderColor", {
        get: function () {
            return this.go;
        },
        set: function (a) {
            this.go = a;
            this.g("BoxTypeFocusBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeBorderColor", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("BoxTypeBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeCornerRadiusBottomRight", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("BoxTypeCornerRadiusBottomRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeCornerRadiusBottomLeft", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("BoxTypeCornerRadiusBottomLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeCornerRadiusTopLeft", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("BoxTypeCornerRadiusTopLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeCornerRadiusTopRight", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("BoxTypeCornerRadiusTopRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeContentPaddingBottom", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("BoxTypeContentPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeContentPaddingLeft", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("BoxTypeContentPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeContentPaddingRight", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("BoxTypeContentPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeContentPaddingTop", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("BoxTypeContentPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeUnderlineWidth", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.g("BoxTypeUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeUnderlineRippleWidth", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("BoxTypeUnderlineRippleWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeHoverUnderlineWidth", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.g("BoxTypeHoverUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeUnderlineColor", {
        get: function () {
            return this.gr;
        },
        set: function (a) {
            this.gr = a;
            this.g("BoxTypeUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeUnderlineRippleColor", {
        get: function () {
            return this.gs;
        },
        set: function (a) {
            this.gs = a;
            this.g("BoxTypeUnderlineRippleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeHoverUnderlineColor", {
        get: function () {
            return this.gq;
        },
        set: function (a) {
            this.gq = a;
            this.g("BoxTypeHoverUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeFocusUnderlineColor", {
        get: function () {
            return this.gp;
        },
        set: function (a) {
            this.gp = a;
            this.g("BoxTypeFocusUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeUnderlineOpacity", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("BoxTypeUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeHoverUnderlineOpacity", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.g("BoxTypeHoverUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeFocusUnderlineOpacity", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.g("BoxTypeFocusUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeFocusUnderlineRippleOpacity", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("BoxTypeFocusUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeUnderlineRippleOpacity", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("BoxTypeUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "boxTypeBackgroundColor", {
        get: function () {
            return this.gm;
        },
        set: function (a) {
            this.gm = a;
            this.g("BoxTypeBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeBorderWidth", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("BorderTypeBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeFocusBorderWidth", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("BorderTypeFocusBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeFocusBorderColor", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("BorderTypeFocusBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeBorderColor", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("BorderTypeBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeCornerRadiusBottomRight", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("BorderTypeCornerRadiusBottomRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeCornerRadiusBottomLeft", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("BorderTypeCornerRadiusBottomLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeCornerRadiusTopLeft", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("BorderTypeCornerRadiusTopLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeCornerRadiusTopRight", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("BorderTypeCornerRadiusTopRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeContentPaddingBottom", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("BorderTypeContentPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeContentPaddingLeft", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("BorderTypeContentPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeContentPaddingRight", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("BorderTypeContentPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeContentPaddingTop", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("BorderTypeContentPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeUnderlineWidth", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("BorderTypeUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeUnderlineRippleWidth", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("BorderTypeUnderlineRippleWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeHoverUnderlineWidth", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("BorderTypeHoverUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeUnderlineColor", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("BorderTypeUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeUnderlineRippleColor", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("BorderTypeUnderlineRippleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeHoverUnderlineColor", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("BorderTypeHoverUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeFocusUnderlineColor", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("BorderTypeFocusUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeUnderlineOpacity", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("BorderTypeUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeHoverUnderlineOpacity", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("BorderTypeHoverUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeFocusUnderlineOpacity", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("BorderTypeFocusUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeFocusUnderlineRippleOpacity", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("BorderTypeFocusUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeUnderlineRippleOpacity", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("BorderTypeUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "borderTypeBackgroundColor", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("BorderTypeBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeBorderWidth", {
        get: function () {
            return this.ch;
        },
        set: function (a) {
            this.ch = a;
            this.g("SearchTypeBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeFocusBorderWidth", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("SearchTypeFocusBorderWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeFocusBorderColor", {
        get: function () {
            return this.g8;
        },
        set: function (a) {
            this.g8 = a;
            this.g("SearchTypeFocusBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeBorderColor", {
        get: function () {
            return this.g7;
        },
        set: function (a) {
            this.g7 = a;
            this.g("SearchTypeBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeCornerRadiusBottomRight", {
        get: function () {
            return this.cn;
        },
        set: function (a) {
            this.cn = a;
            this.g("SearchTypeCornerRadiusBottomRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeCornerRadiusBottomLeft", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.g("SearchTypeCornerRadiusBottomLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeCornerRadiusTopLeft", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.g("SearchTypeCornerRadiusTopLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeCornerRadiusTopRight", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("SearchTypeCornerRadiusTopRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeContentPaddingBottom", {
        get: function () {
            return this.ci;
        },
        set: function (a) {
            this.ci = a;
            this.g("SearchTypeContentPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeContentPaddingLeft", {
        get: function () {
            return this.cj;
        },
        set: function (a) {
            this.cj = a;
            this.g("SearchTypeContentPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeContentPaddingRight", {
        get: function () {
            return this.ck;
        },
        set: function (a) {
            this.ck = a;
            this.g("SearchTypeContentPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeContentPaddingTop", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            this.cl = a;
            this.g("SearchTypeContentPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeUnderlineWidth", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("SearchTypeUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeUnderlineRippleWidth", {
        get: function () {
            return this.cx;
        },
        set: function (a) {
            this.cx = a;
            this.g("SearchTypeUnderlineRippleWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeHoverUnderlineWidth", {
        get: function () {
            return this.cu;
        },
        set: function (a) {
            this.cu = a;
            this.g("SearchTypeHoverUnderlineWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeUnderlineColor", {
        get: function () {
            return this.hb;
        },
        set: function (a) {
            this.hb = a;
            this.g("SearchTypeUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeUnderlineRippleColor", {
        get: function () {
            return this.hc;
        },
        set: function (a) {
            this.hc = a;
            this.g("SearchTypeUnderlineRippleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeHoverUnderlineColor", {
        get: function () {
            return this.ha;
        },
        set: function (a) {
            this.ha = a;
            this.g("SearchTypeHoverUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeFocusUnderlineColor", {
        get: function () {
            return this.g9;
        },
        set: function (a) {
            this.g9 = a;
            this.g("SearchTypeFocusUnderlineColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeUnderlineOpacity", {
        get: function () {
            return this.cv;
        },
        set: function (a) {
            this.cv = a;
            this.g("SearchTypeUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeHoverUnderlineOpacity", {
        get: function () {
            return this.ct;
        },
        set: function (a) {
            this.ct = a;
            this.g("SearchTypeHoverUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeFocusUnderlineOpacity", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("SearchTypeFocusUnderlineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeFocusUnderlineRippleOpacity", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("SearchTypeFocusUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeUnderlineRippleOpacity", {
        get: function () {
            return this.cw;
        },
        set: function (a) {
            this.cw = a;
            this.g("SearchTypeUnderlineRippleOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "searchTypeBackgroundColor", {
        get: function () {
            return this.g6;
        },
        set: function (a) {
            this.g6 = a;
            this.g("SearchTypeBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "isExpanded", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IsExpanded");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "isHovered", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("IsHovered");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "isFocused", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("IsFocused");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "actualIsExpanded", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ActualIsExpanded");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "focusBorderColor", {
        get: function () {
            return this.gv;
        },
        set: function (a) {
            this.gv = a;
            this.g("FocusBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "id", {
        get: function () {
            return this.gy;
        },
        set: function (a) {
            this.gy = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "disabled", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputGroupDescription.prototype, "inputs", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Inputs");
        },
        enumerable: false,
        configurable: true
    });
    InputGroupDescription.$t = markType(InputGroupDescription, 'InputGroupDescription', Description.$);
    return InputGroupDescription;
}(Description));
export { InputGroupDescription };
