// igniteui-react-core/esm5/lib/InputDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { InputGroupItemDescription } from "./InputGroupItemDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var InputDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(InputDescription, _super);
    function InputDescription() {
        var _this = _super.call(this) || this;
        _this.ag = null;
        _this.ak = null;
        _this.ac = null;
        _this.av = null;
        _this.ax = null;
        _this.l = false;
        _this.o = false;
        _this.ae = null;
        _this.ad = null;
        _this.az = null;
        _this.al = null;
        _this.u = 0;
        _this.am = null;
        _this.an = null;
        _this.ar = null;
        _this.aw = null;
        _this.ap = null;
        _this.aq = null;
        _this.y = 0;
        _this.ao = null;
        _this.af = null;
        _this.m = false;
        _this.a0 = null;
        _this.k = false;
        _this.n = false;
        _this.ay = null;
        _this.x = 0;
        _this.w = 0;
        _this.as = null;
        _this.au = null;
        _this.at = null;
        _this.ah = null;
        _this.ai = null;
        _this.aj = null;
        return _this;
    }
    InputDescription.prototype.get_type = function () {
        return "Input";
    };
    Object.defineProperty(InputDescription.prototype, "baseTheme", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "density", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("Density");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "actualDensity", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("ActualDensity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "mask", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("Mask");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "promptChar", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("PromptChar");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "includeLiterals", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IncludeLiterals");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "showSpinner", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("ShowSpinner");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "actualTextColor", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("ActualTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "actualHoverTextColor", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("ActualHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "textColor", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "fontFamily", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("FontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "fontSize", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("FontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "fontStyle", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("FontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "fontWeight", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("FontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "inputType", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("InputType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "placeholder", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("Placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "hoverTextColor", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("HoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "id", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "tabIndex", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("TabIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "$for", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("For");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "ariaLabel", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("AriaLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "isHover", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("IsHover");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "value", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "disabled", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "readonly", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Readonly");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "textAlignment", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("TextAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "selectionStart", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("SelectionStart");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "selectionEnd", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("SelectionEnd");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "keyDownRef", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("KeyDownRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "keyUpRef", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("KeyUpRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "keyPressRef", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("KeyPressRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "changeRef", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "changingRef", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("ChangingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(InputDescription.prototype, "compositionEndRef", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("CompositionEndRef");
        },
        enumerable: false,
        configurable: true
    });
    InputDescription.$t = markType(InputDescription, 'InputDescription', InputGroupItemDescription.$);
    return InputDescription;
}(InputGroupItemDescription));
export { InputDescription };
