// igniteui-react-core/esm5/lib/ItemLegendDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { LegendBaseDescription } from "./LegendBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ItemLegendDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ItemLegendDescription, _super);
    function ItemLegendDescription() {
        var _this = _super.call(this) || this;
        _this.y = null;
        _this.u = null;
        _this.v = null;
        _this.w = null;
        _this.x = null;
        return _this;
    }
    Object.defineProperty(ItemLegendDescription.prototype, "width", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ItemLegendDescription.prototype, "height", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    ItemLegendDescription.prototype.get_type = function () {
        return "ItemLegend";
    };
    Object.defineProperty(ItemLegendDescription.prototype, "orientation", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("Orientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ItemLegendDescription.prototype, "textColor", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ItemLegendDescription.prototype, "textStyle", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("TextStyle");
        },
        enumerable: false,
        configurable: true
    });
    ItemLegendDescription.$t = markType(ItemLegendDescription, 'ItemLegendDescription', LegendBaseDescription.$);
    return ItemLegendDescription;
}(LegendBaseDescription));
export { ItemLegendDescription };
