// igniteui-react-core/esm5/lib/ItemToolTipLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ItemToolTipLayerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ItemToolTipLayerDescription, _super);
    function ItemToolTipLayerDescription() {
        var _this = _super.call(this) || this;
        _this.gd = null;
        _this.ge = null;
        _this.ga = false;
        _this.f9 = false;
        return _this;
    }
    ItemToolTipLayerDescription.prototype.get_type = function () {
        return "ItemToolTipLayer";
    };
    Object.defineProperty(ItemToolTipLayerDescription.prototype, "targetSeriesName", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("TargetSeriesName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ItemToolTipLayerDescription.prototype, "targetSeriesRef", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("TargetSeriesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ItemToolTipLayerDescription.prototype, "useInterpolation", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("UseInterpolation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ItemToolTipLayerDescription.prototype, "skipUnknownValues", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("SkipUnknownValues");
        },
        enumerable: false,
        configurable: true
    });
    ItemToolTipLayerDescription.$t = markType(ItemToolTipLayerDescription, 'ItemToolTipLayerDescription', AnnotationLayerDescription.$);
    return ItemToolTipLayerDescription;
}(AnnotationLayerDescription));
export { ItemToolTipLayerDescription };
