// igniteui-react-core/esm5/lib/JsonWriter.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, markType } from "./type";
import { StringBuilder } from "./StringBuilder";
/**
 * @hidden
 */
var JsonWriter = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(JsonWriter, _super);
    function JsonWriter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.d = new StringBuilder(0);
        _this.b = 0;
        _this.a = true;
        return _this;
    }
    JsonWriter.prototype.g = function () {
        this.b++;
    };
    JsonWriter.prototype.e = function () {
        this.b--;
    };
    JsonWriter.prototype.i = function (a) {
        this.f();
        this.d.l(a);
    };
    JsonWriter.prototype.f = function () {
        if (this.a) {
            this.a = false;
            for (var a = 0; a < this.b; a++) {
                this.d.l("\t");
            }
        }
    };
    JsonWriter.prototype.j = function (a) {
        this.f();
        this.d.u(a);
        this.a = true;
    };
    JsonWriter.prototype.h = function () {
        this.f();
        this.d.t();
        this.a = true;
    };
    JsonWriter.prototype.toString = function () {
        return this.d.toString();
    };
    JsonWriter.$t = markType(JsonWriter, 'JsonWriter');
    return JsonWriter;
}(Base));
export { JsonWriter };
