// igniteui-react-core/esm5/lib/LabelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { InputGroupItemDescription } from "./InputGroupItemDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var LabelDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(LabelDescription, _super);
    function LabelDescription() {
        var _this = _super.call(this) || this;
        _this.ac = null;
        _this.ad = null;
        _this.u = null;
        _this.y = null;
        _this.v = null;
        _this.w = null;
        _this.x = null;
        _this.aq = null;
        _this.al = null;
        _this.am = null;
        _this.ah = null;
        _this.q = 0;
        _this.ai = null;
        _this.aj = null;
        _this.an = null;
        _this.ao = null;
        _this.ae = null;
        _this.af = null;
        _this.ag = null;
        _this.z = null;
        _this.aa = null;
        _this.s = 0;
        _this.ak = null;
        _this.ab = null;
        _this.ap = null;
        _this.l = false;
        _this.m = false;
        _this.k = false;
        return _this;
    }
    LabelDescription.prototype.get_type = function () {
        return "Label";
    };
    Object.defineProperty(LabelDescription.prototype, "baseTheme", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "density", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("Density");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "actualDensity", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("ActualDensity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "actualTextColor", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("ActualTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "actualHighlightTextColor", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("ActualHighlightTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "actualHoverHighlightTextColor", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("ActualHoverHighlightTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "actualHoverTextColor", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("ActualHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "textColor", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "highlightTextColor", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("HighlightTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "hoverHighlightTextColor", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("HoverHighlightTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "fontFamily", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("FontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "fontSize", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("FontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "fontStyle", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("FontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "fontWeight", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("FontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "hoverTextColor", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("HoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "id", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "display", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("Display");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "flexDirection", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("FlexDirection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "flexGrow", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("FlexGrow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "alignItems", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("AlignItems");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "alignSelf", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("AlignSelf");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "tabIndex", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("TabIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "$for", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("For");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "ariaLabel", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("AriaLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "text", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("Text");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "isHover", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IsHover");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "value", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LabelDescription.prototype, "disabled", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    LabelDescription.$t = markType(LabelDescription, 'LabelDescription', InputGroupItemDescription.$);
    return LabelDescription;
}(InputGroupItemDescription));
export { LabelDescription };
