// igniteui-react-core/esm5/lib/LegendBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var LegendBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(LegendBaseDescription, _super);
    function LegendBaseDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.k = null;
        _this.l = null;
        _this.i = null;
        _this.j = null;
        _this.m = null;
        return _this;
    }
    Object.defineProperty(LegendBaseDescription.prototype, "background", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    LegendBaseDescription.prototype.get_type = function () {
        return "LegendBase";
    };
    Object.defineProperty(LegendBaseDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LegendBaseDescription.prototype, "legendItemMouseLeftButtonDownRef", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("LegendItemMouseLeftButtonDownRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LegendBaseDescription.prototype, "legendItemMouseLeftButtonUpRef", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("LegendItemMouseLeftButtonUpRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LegendBaseDescription.prototype, "legendItemMouseEnterRef", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("LegendItemMouseEnterRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LegendBaseDescription.prototype, "legendItemMouseLeaveRef", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("LegendItemMouseLeaveRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LegendBaseDescription.prototype, "legendItemMouseMoveRef", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("LegendItemMouseMoveRef");
        },
        enumerable: false,
        configurable: true
    });
    LegendBaseDescription.$t = markType(LegendBaseDescription, 'LegendBaseDescription', Description.$);
    return LegendBaseDescription;
}(Description));
export { LegendBaseDescription };
