// igniteui-react-core/esm5/lib/LinearGraphRangeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var LinearGraphRangeDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(LinearGraphRangeDescription, _super);
    function LinearGraphRangeDescription() {
        var _this = _super.call(this) || this;
        _this.w = null;
        _this.v = null;
        _this.x = null;
        _this.m = 0;
        _this.h = 0;
        _this.j = 0;
        _this.i = 0;
        _this.l = 0;
        _this.k = 0;
        _this.n = 0;
        return _this;
    }
    LinearGraphRangeDescription.prototype.get_type = function () {
        return "LinearGraphRange";
    };
    Object.defineProperty(LinearGraphRangeDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LinearGraphRangeDescription.prototype, "name", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LinearGraphRangeDescription.prototype, "brush", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("Brush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LinearGraphRangeDescription.prototype, "outline", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("Outline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LinearGraphRangeDescription.prototype, "startValue", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("StartValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LinearGraphRangeDescription.prototype, "endValue", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("EndValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LinearGraphRangeDescription.prototype, "innerStartExtent", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("InnerStartExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LinearGraphRangeDescription.prototype, "innerEndExtent", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("InnerEndExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LinearGraphRangeDescription.prototype, "outerStartExtent", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("OuterStartExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LinearGraphRangeDescription.prototype, "outerEndExtent", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("OuterEndExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LinearGraphRangeDescription.prototype, "strokeThickness", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("StrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    LinearGraphRangeDescription.$t = markType(LinearGraphRangeDescription, 'LinearGraphRangeDescription', Description.$);
    return LinearGraphRangeDescription;
}(Description));
export { LinearGraphRangeDescription };
