// igniteui-react-core/esm5/lib/ListPanelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ListPanelDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ListPanelDescription, _super);
    function ListPanelDescription() {
        var _this = _super.call(this) || this;
        _this.bc = null;
        _this.t = false;
        _this.a7 = null;
        _this.a2 = null;
        _this.ap = null;
        _this.ae = 0;
        _this.ag = 0;
        _this.af = 0;
        _this.ad = 0;
        _this.aq = null;
        _this.ba = null;
        _this.i = null;
        _this.h = null;
        _this.k = null;
        _this.l = null;
        _this.ay = null;
        _this.r = null;
        _this.j = null;
        _this.ai = 0;
        _this.z = 0;
        _this.au = null;
        _this.aa = 0;
        _this.av = null;
        _this.aw = null;
        _this.a6 = null;
        _this.a5 = null;
        _this.at = null;
        _this.a3 = null;
        _this.u = false;
        _this.bd = null;
        _this.ah = 0;
        _this.v = false;
        _this.a8 = null;
        _this.a9 = null;
        _this.bb = null;
        _this.ar = null;
        _this.ax = null;
        _this.a4 = null;
        _this.a1 = null;
        _this.a0 = null;
        _this.az = null;
        _this.as = null;
        return _this;
    }
    ListPanelDescription.prototype.get_type = function () {
        return "ListPanel";
    };
    Object.defineProperty(ListPanelDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "selectionMode", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("SelectionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "isActiveRowStyleEnabled", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("IsActiveRowStyleEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "selectedBackground", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("SelectedBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "normalBackground", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("NormalBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "activationBorder", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("ActivationBorder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "activationBorderLeftWidth", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("ActivationBorderLeftWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "activationBorderTopWidth", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("ActivationBorderTopWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "activationBorderRightWidth", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("ActivationBorderRightWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "activationBorderBottomWidth", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("ActivationBorderBottomWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "activationMode", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("ActivationMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "selectionBehavior", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("SelectionBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "primaryKey", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("PrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "actualPrimaryKey", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ActualPrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "selectedItems", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("SelectedItems");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "selectedKeys", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("SelectedKeys");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "dataSource", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("DataSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "schemaIncludedProperties", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("SchemaIncludedProperties");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "rowHeight", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("RowHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "actualRowHeight", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("ActualRowHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "cellFontFamily", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("CellFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "cellFontSize", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("CellFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "cellFontStyle", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("CellFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "cellFontWeight", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("CellFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "scrollbarStyle", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("ScrollbarStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "scrollbarBackground", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("ScrollbarBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "backgroundColor", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("BackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "orientation", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("Orientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "isCustomRowHeightEnabled", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("IsCustomRowHeightEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "textColor", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "activeRow", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("ActiveRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "notifyOnAllSelectionChanges", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("NotifyOnAllSelectionChanges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "selectedItemsChangedRef", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("SelectedItemsChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "selectedKeysChangedRef", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("SelectedKeysChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "selectionChangedRef", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("SelectionChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "activeRowChangedRef", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("ActiveRowChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "contentRefreshedRef", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("ContentRefreshedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "rowUpdatingRef", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("RowUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "itemWidthRequestedRef", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("ItemWidthRequestedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "itemHeightRequestedRef", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("ItemHeightRequestedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "itemClickedRef", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("ItemClickedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelDescription.prototype, "actualPrimaryKeyChangeRef", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("ActualPrimaryKeyChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    ListPanelDescription.$t = markType(ListPanelDescription, 'ListPanelDescription', Description.$);
    return ListPanelDescription;
}(Description));
export { ListPanelDescription };
