// igniteui-react-core/esm5/lib/ListPanelItemModelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ListPanelItemModelDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ListPanelItemModelDescription, _super);
    function ListPanelItemModelDescription() {
        var _this = _super.call(this) || this;
        _this.v = 0;
        _this.q = 0;
        _this.x = null;
        _this.j = false;
        _this.p = 0;
        _this.r = 0;
        _this.h = false;
        _this.i = false;
        _this.k = false;
        return _this;
    }
    ListPanelItemModelDescription.prototype.get_type = function () {
        return "ListPanelItemModel";
    };
    Object.defineProperty(ListPanelItemModelDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemModelDescription.prototype, "dataRow", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("DataRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemModelDescription.prototype, "rowHeight", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("RowHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemModelDescription.prototype, "rowObjectRef", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("RowObjectRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemModelDescription.prototype, "isModelDirty", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("IsModelDirty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemModelDescription.prototype, "left", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Left");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemModelDescription.prototype, "top", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Top");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemModelDescription.prototype, "isActivated", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("IsActivated");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemModelDescription.prototype, "isActivationSupported", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("IsActivationSupported");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemModelDescription.prototype, "isSelected", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("IsSelected");
        },
        enumerable: false,
        configurable: true
    });
    ListPanelItemModelDescription.$t = markType(ListPanelItemModelDescription, 'ListPanelItemModelDescription', Description.$);
    return ListPanelItemModelDescription;
}(Description));
export { ListPanelItemModelDescription };
