// igniteui-react-core/esm5/lib/ListPanelItemEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ListPanelItemEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ListPanelItemEventArgsDescription, _super);
    function ListPanelItemEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.h = null;
        _this.m = false;
        _this.l = false;
        return _this;
    }
    ListPanelItemEventArgsDescription.prototype.get_type = function () {
        return "ListPanelItemEventArgs";
    };
    Object.defineProperty(ListPanelItemEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemEventArgsDescription.prototype, "itemInfo", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ItemInfo");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemEventArgsDescription.prototype, "listPanel", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ListPanel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemEventArgsDescription.prototype, "isLeftButton", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("IsLeftButton");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ListPanelItemEventArgsDescription.prototype, "isDoubleClick", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IsDoubleClick");
        },
        enumerable: false,
        configurable: true
    });
    ListPanelItemEventArgsDescription.$t = markType(ListPanelItemEventArgsDescription, 'ListPanelItemEventArgsDescription', Description.$);
    return ListPanelItemEventArgsDescription;
}(Description));
export { ListPanelItemEventArgsDescription };
