// igniteui-react-core/esm5/lib/NativeUIGrid.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { NativeUIContainer } from "./NativeUIContainer";
import { NativeUIBoxSizing_$type } from "./NativeUIBoxSizing";
import { enumGetBox, EnumUtil, markType } from "./type";
import { NativeUIPlaceContent_$type } from "./NativeUIPlaceContent";
/**
 * @hidden
 */
var NativeUIGrid = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NativeUIGrid, _super);
    function NativeUIGrid() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.bt = null;
        return _this;
    }
    NativeUIGrid.prototype.get_e = function () {
        return 7;
    };
    Object.defineProperty(NativeUIGrid.prototype, "b1", {
        get: function () {
            return this.br.getRowDefinitionsCount(this.am);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIGrid.prototype, "bx", {
        get: function () {
            return this.br.getColumnDefinitionsCount(this.am);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIGrid.prototype, "br", {
        get: function () {
            return this.d;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIGrid.prototype, "bs", {
        get: function () {
            var a = this.ag(38);
            return EnumUtil.getEnumValue(NativeUIBoxSizing_$type, a);
        },
        set: function (a) {
            this.a4(38, enumGetBox(NativeUIBoxSizing_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIGrid.prototype, "bv", {
        get: function () {
            var a = this.ag(72);
            return EnumUtil.getEnumValue(NativeUIPlaceContent_$type, a);
        },
        set: function (a) {
            this.a4(72, enumGetBox(NativeUIPlaceContent_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIGrid.prototype, "bw", {
        get: function () {
            return this.ag(37);
        },
        set: function (a) {
            this.a4(37, a);
        },
        enumerable: false,
        configurable: true
    });
    NativeUIGrid.prototype.b4 = function (a, b, c) {
        this.br.addRowDefinition(this.am, a, b, c);
    };
    NativeUIGrid.prototype.b5 = function (a, b, c, d) {
        this.br.addRowDefinitionWithMin(this.am, a, b, c, d);
    };
    NativeUIGrid.prototype.ce = function (a, b, c) {
        this.br.updateRowDefinition(this.am, a, b, c);
    };
    NativeUIGrid.prototype.b2 = function (a, b, c) {
        this.br.addColumnDefinition(this.am, a, b, c);
    };
    NativeUIGrid.prototype.b3 = function (a, b, c, d) {
        this.br.addColumnDefinitionWithMin(this.am, a, b, c, d);
    };
    NativeUIGrid.prototype.cd = function (a, b, c) {
        this.br.updateColumnDefinition(this.am, a, b, c);
    };
    NativeUIGrid.prototype.b6 = function (a, b, c) {
        this.br.insertColumnDefinition(this.am, a, b, c);
    };
    NativeUIGrid.prototype.b7 = function (a, b, c, d) {
        this.br.insertColumnDefinitionWithMin(this.am, a, b, c, d);
    };
    NativeUIGrid.prototype.cc = function (a, b) {
        this.br.setRow(this.am, a, b);
    };
    NativeUIGrid.prototype.ca = function (a, b) {
        this.br.setColumn(this.am, a, b);
    };
    NativeUIGrid.prototype.cb = function (a, b) {
        this.br.setColumnSpan(this.am, a, b);
    };
    NativeUIGrid.prototype.b0 = function (a) {
        return this.br.getRow(this.am, a);
    };
    NativeUIGrid.prototype.by = function (a) {
        return this.br.getColumn(this.am, a);
    };
    NativeUIGrid.prototype.bz = function (a) {
        return this.br.getColumnSpan(this.am, a);
    };
    NativeUIGrid.prototype.b9 = function (a) {
        this.br.removeRowDefinition(this.am, a);
    };
    NativeUIGrid.prototype.b8 = function (a) {
        this.br.removeColumnDefinition(this.am, a);
    };
    Object.defineProperty(NativeUIGrid.prototype, "bu", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            var b = this.bt;
            this.bt = a;
            if (this.bt != b) {
                if (b != null) {
                    this.a2(1, b);
                }
                if (this.bt != null) {
                    this.aq(1, this.bt);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    NativeUIGrid.$t = markType(NativeUIGrid, 'NativeUIGrid', NativeUIContainer.$);
    return NativeUIGrid;
}(NativeUIContainer));
export { NativeUIGrid };
