// igniteui-react-core/esm5/lib/NativeUICheckboxList.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { NativeUIComponent } from "./NativeUIComponent";
import { enumGetBox, EnumUtil, markType } from "./type";
import { NativeUICheckboxListIndexType_$type } from "./NativeUICheckboxListIndexType";
import { ControlDisplayDensity_$type } from "./ControlDisplayDensity";
/**
 * @hidden
 */
var NativeUICheckboxList = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NativeUICheckboxList, _super);
    function NativeUICheckboxList() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.bn = null;
        _this.bo = null;
        _this.bp = null;
        _this.bu = null;
        _this.bw = null;
        return _this;
    }
    NativeUICheckboxList.prototype.get_e = function () {
        return 16;
    };
    Object.defineProperty(NativeUICheckboxList.prototype, "bq", {
        get: function () {
            return this.bn;
        },
        set: function (a) {
            var b = this.bn;
            this.bn = a;
            if (this.bn != b) {
                if (b != null) {
                    this.a2(2, b);
                }
                if (this.bn != null) {
                    this.aq(2, this.bn);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "br", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            var b = this.bo;
            this.bo = a;
            if (this.bo != b) {
                if (b != null) {
                    this.a2(18, b);
                }
                if (this.bo != null) {
                    this.aq(18, this.bo);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "bs", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            var b = this.bp;
            this.bp = a;
            if (this.bp != b) {
                if (b != null) {
                    this.a2(19, b);
                }
                if (this.bp != null) {
                    this.aq(19, this.bp);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "bv", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            var b = this.bu;
            this.bu = a;
            if (this.bu != b) {
                if (b != null) {
                    this.a2(21, b);
                }
                if (this.bp != null) {
                    this.aq(21, this.bu);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "bx", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            var b = this.bw;
            this.bw = a;
            if (this.bw != b) {
                if (b != null) {
                    this.a2(25, b);
                }
                if (this.bp != null) {
                    this.aq(25, this.bw);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "bz", {
        get: function () {
            var a = this.ag(8);
            return a;
        },
        set: function (a) {
            this.a4(8, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "b6", {
        get: function () {
            var a = this.ag(23);
            return a;
        },
        set: function (a) {
            this.a4(23, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "b0", {
        get: function () {
            var a = this.ag(75);
            return a;
        },
        set: function (a) {
            this.a4(75, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "b1", {
        get: function () {
            var a = this.ag(74);
            return a;
        },
        set: function (a) {
            this.a4(74, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "bt", {
        get: function () {
            var a = this.ag(96);
            return EnumUtil.getEnumValue(NativeUICheckboxListIndexType_$type, a);
        },
        set: function (a) {
            this.a4(96, enumGetBox(NativeUICheckboxListIndexType_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "bl", {
        get: function () {
            var a = this.ag(1);
            return EnumUtil.getEnumValue(ControlDisplayDensity_$type, a);
        },
        set: function (a) {
            this.a4(1, enumGetBox(ControlDisplayDensity_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "by", {
        get: function () {
            var a = this.ag(76);
            return a;
        },
        set: function (a) {
            this.a4(76, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "bk", {
        get: function () {
            var a = this.ag(77);
            return a;
        },
        set: function (a) {
            this.a4(77, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "ca", {
        get: function () {
            var a = this.ag(33);
            return a;
        },
        set: function (a) {
            this.a4(33, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "b7", {
        get: function () {
            var a = this.ag(24);
            return a;
        },
        set: function (a) {
            this.a4(24, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "b8", {
        get: function () {
            var a = this.ag(78);
            return a;
        },
        set: function (a) {
            this.a4(78, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "b9", {
        get: function () {
            var a = this.ag(79);
            return a;
        },
        set: function (a) {
            this.a4(79, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUICheckboxList.prototype, "bm", {
        get: function () {
            return this.d;
        },
        enumerable: false,
        configurable: true
    });
    NativeUICheckboxList.prototype.b3 = function () {
        this.bm.clearSelectedKeys(this.am);
    };
    NativeUICheckboxList.prototype.b2 = function (a, b) {
        this.bm.addSelectedKey(this.am, a, b);
    };
    NativeUICheckboxList.prototype.b4 = function (a, b, c) {
        this.bm.insertSelectedKey(this.am, a, b, c);
    };
    NativeUICheckboxList.prototype.b5 = function (a) {
        this.bm.removeSelectedKeyAt(this.am, a);
    };
    NativeUICheckboxList.$t = markType(NativeUICheckboxList, 'NativeUICheckboxList', NativeUIComponent.$);
    return NativeUICheckboxList;
}(NativeUIComponent));
export { NativeUICheckboxList };
