// igniteui-react-core/esm5/lib/NativeUIInput.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIInputDisplayType_$type } from "./NativeUIInputDisplayType";
import { enumGetBox, EnumUtil, typeGetValue, markType } from "./type";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { NativeUIInputType_$type } from "./NativeUIInputType";
import { NativeUIComponentHorizontalAlignment_$type } from "./NativeUIComponentHorizontalAlignment";
/**
 * @hidden
 */
var NativeUIInput = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NativeUIInput, _super);
    function NativeUIInput() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.bq = null;
        _this.bw = null;
        _this.by = null;
        _this.br = null;
        _this.bm = null;
        _this.bk = null;
        return _this;
    }
    Object.defineProperty(NativeUIInput.prototype, "bl", {
        get: function () {
            return this.d;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "bu", {
        get: function () {
            var a = this.ag(0);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIInputDisplayType_$type, a);
        },
        set: function (a) {
            this.a4(0, enumGetBox(NativeUIInputDisplayType_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "bp", {
        get: function () {
            var a = this.ag(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        },
        set: function (a) {
            this.a4(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "b5", {
        get: function () {
            var a = this.ag(3);
            return a;
        },
        set: function (a) {
            this.a4(3, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "cb", {
        get: function () {
            var a = this.ag(100);
            return a;
        },
        set: function (a) {
            this.a4(100, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "b7", {
        get: function () {
            var a = this.ag(2);
            return a;
        },
        set: function (a) {
            this.a4(2, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "bs", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            var b = this.bq;
            this.bq = a;
            if (this.bq != b) {
                if (b != null) {
                    this.a2(0, b);
                }
                if (this.bq != null) {
                    this.aq(0, this.bq);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "bx", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            var b = this.bw;
            this.bw = a;
            if (this.bw != b) {
                if (b != null) {
                    this.a2(27, b);
                }
                if (this.bw != null) {
                    this.aq(27, this.bw);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "bz", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            var b = this.by;
            this.by = a;
            if (this.by != b) {
                if (b != null) {
                    this.a2(9, b);
                }
                if (this.by != null) {
                    this.aq(9, this.by);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "bt", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            var b = this.br;
            this.br = a;
            if (this.br != b) {
                if (b != null) {
                    this.a2(26, b);
                }
                if (this.bq != null) {
                    this.aq(26, this.bq);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "bn", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            var b = this.bm;
            this.bm = a;
            if (this.bm != b) {
                if (b != null) {
                    this.a2(24, b);
                }
                if (this.bm != null) {
                    this.aq(24, this.bm);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "bv", {
        get: function () {
            var a = this.ag(19);
            return a == null ? 0 : EnumUtil.getEnumValue(NativeUIInputType_$type, a);
        },
        set: function (a) {
            this.a4(19, enumGetBox(NativeUIInputType_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "bo", {
        get: function () {
            var a = this.ag(20);
            return a == null ? 0 : EnumUtil.getEnumValue(NativeUIComponentHorizontalAlignment_$type, a);
        },
        set: function (a) {
            this.a4(20, enumGetBox(NativeUIComponentHorizontalAlignment_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "ce", {
        get: function () {
            var a = this.ag(33);
            return a == null ? null : a;
        },
        set: function (a) {
            this.a4(33, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "b2", {
        get: function () {
            var a = this.ag(21);
            return a == null ? false : a;
        },
        set: function (a) {
            this.a4(21, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "b0", {
        get: function () {
            var a = this.ag(32);
            return a == null ? false : a;
        },
        set: function (a) {
            this.a4(32, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "b6", {
        get: function () {
            var a = this.ag(97);
            return a;
        },
        set: function (a) {
            this.a4(97, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "cc", {
        get: function () {
            var a = this.ag(98);
            return a;
        },
        set: function (a) {
            this.a4(98, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "cd", {
        get: function () {
            var a = this.ag(99);
            return a;
        },
        set: function (a) {
            this.a4(99, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "b4", {
        get: function () {
            var a = this.ag(102);
            return typeGetValue(a);
        },
        set: function (a) {
            this.a4(102, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "b3", {
        get: function () {
            var a = this.ag(103);
            return typeGetValue(a);
        },
        set: function (a) {
            this.a4(103, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIInput.prototype, "b1", {
        get: function () {
            var a = this.ag(104);
            return a;
        },
        set: function (a) {
            this.a4(104, a);
        },
        enumerable: false,
        configurable: true
    });
    NativeUIInput.prototype.b8 = function () {
        this.bl.resetIsExpanded(this.am);
    };
    NativeUIInput.prototype.b9 = function () {
        this.bl.select(this.am);
    };
    NativeUIInput.prototype.ca = function (a, b) {
        this.bl.setSelectionRange(this.am, a, b);
    };
    NativeUIInput.$t = markType(NativeUIInput, 'NativeUIInput', NativeUIComponent.$);
    return NativeUIInput;
}(NativeUIComponent));
export { NativeUIInput };
