// igniteui-react-core/esm5/lib/OperationFilterExpressionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { FilterExpressionDescription } from "./FilterExpressionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var OperationFilterExpressionDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(OperationFilterExpressionDescription, _super);
    function OperationFilterExpressionDescription() {
        var _this = _super.call(this) || this;
        _this.r = null;
        _this.s = null;
        _this.k = null;
        _this.l = null;
        _this.q = null;
        return _this;
    }
    OperationFilterExpressionDescription.prototype.get_type = function () {
        return "OperationFilterExpression";
    };
    Object.defineProperty(OperationFilterExpressionDescription.prototype, "propertyName", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("PropertyName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(OperationFilterExpressionDescription.prototype, "valueRef", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("ValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(OperationFilterExpressionDescription.prototype, "left", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Left");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(OperationFilterExpressionDescription.prototype, "right", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Right");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(OperationFilterExpressionDescription.prototype, "operator", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Operator");
        },
        enumerable: false,
        configurable: true
    });
    OperationFilterExpressionDescription.$t = markType(OperationFilterExpressionDescription, 'OperationFilterExpressionDescription', FilterExpressionDescription.$);
    OperationFilterExpressionDescription.__marshalByValue1 = true;
    OperationFilterExpressionDescription.__marshalByValueAlias1 = "OperationFilterExpression";
    return OperationFilterExpressionDescription;
}(FilterExpressionDescription));
export { OperationFilterExpressionDescription };
