// igniteui-react-core/esm5/lib/PlotAreaMouseButtonEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var PlotAreaMouseButtonEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PlotAreaMouseButtonEventArgsDescription, _super);
    function PlotAreaMouseButtonEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.i = null;
        _this.n = false;
        _this.l = null;
        return _this;
    }
    PlotAreaMouseButtonEventArgsDescription.prototype.get_type = function () {
        return "PlotAreaMouseButtonEventArgs";
    };
    Object.defineProperty(PlotAreaMouseButtonEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PlotAreaMouseButtonEventArgsDescription.prototype, "chartPosition", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ChartPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PlotAreaMouseButtonEventArgsDescription.prototype, "plotAreaPosition", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("PlotAreaPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PlotAreaMouseButtonEventArgsDescription.prototype, "manipulationOccurred", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ManipulationOccurred");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PlotAreaMouseButtonEventArgsDescription.prototype, "viewer", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Viewer");
        },
        enumerable: false,
        configurable: true
    });
    PlotAreaMouseButtonEventArgsDescription.$t = markType(PlotAreaMouseButtonEventArgsDescription, 'PlotAreaMouseButtonEventArgsDescription', Description.$);
    return PlotAreaMouseButtonEventArgsDescription;
}(Description));
export { PlotAreaMouseButtonEventArgsDescription };
