// igniteui-react-core/esm5/lib/PlotAreaMouseButtonEventArgsDescriptionMetadata_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, String_$type, markType } from "./type";
import { Dictionary$2 } from "./Dictionary$2";
import { PlotAreaMouseButtonEventArgsDescription } from "./PlotAreaMouseButtonEventArgsDescription";
import { PlotAreaMouseEventArgsDescription } from "./PlotAreaMouseEventArgsDescription";
import { SeriesDescriptionMetadata } from "./SeriesDescriptionMetadata";
import { ChartCursorEventArgsDescription } from "./ChartCursorEventArgsDescription";
import { ChartMouseEventArgsDescription } from "./ChartMouseEventArgsDescription";
import { DataChartMouseButtonEventArgsDescription } from "./DataChartMouseButtonEventArgsDescription";
import { SeriesViewerManipulationEventArgsDescriptionMetadata } from "./SeriesViewerManipulationEventArgsDescriptionMetadata";
import { RectChangedEventArgsDescriptionMetadata } from "./RectChangedEventArgsDescriptionMetadata";
import { SeriesViewerSelectedSeriesItemsChangedEventArgsDescriptionMetadata } from "./SeriesViewerSelectedSeriesItemsChangedEventArgsDescriptionMetadata";
import { RefreshCompletedEventArgsDescriptionMetadata } from "./RefreshCompletedEventArgsDescriptionMetadata";
import { ImageCapturedEventArgsDescriptionMetadata } from "./ImageCapturedEventArgsDescriptionMetadata";
/**
 * @hidden
 */
var PlotAreaMouseButtonEventArgsDescriptionMetadata = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PlotAreaMouseButtonEventArgsDescriptionMetadata, _super);
    function PlotAreaMouseButtonEventArgsDescriptionMetadata() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PlotAreaMouseButtonEventArgsDescriptionMetadata.b = function (a) {
        if (PlotAreaMouseButtonEventArgsDescriptionMetadata.a == null) {
            PlotAreaMouseButtonEventArgsDescriptionMetadata.a = new Dictionary$2(String_$type, String_$type, 0);
            PlotAreaMouseButtonEventArgsDescriptionMetadata.c(PlotAreaMouseButtonEventArgsDescriptionMetadata.a);
        }
        if (a.k(PlotAreaMouseButtonEventArgsDescriptionMetadata.a)) {
            return;
        }
        a.ac(PlotAreaMouseButtonEventArgsDescriptionMetadata.a);
        SeriesViewerDescriptionMetadata.d(a);
    };
    PlotAreaMouseButtonEventArgsDescriptionMetadata.c = function (a) {
        a.item("ChartPosition", "Point");
        a.item("PlotAreaPosition", "Point");
        a.item("ManipulationOccurred", "Boolean");
        a.item("Viewer", "ExportedType:SeriesViewer");
    };
    PlotAreaMouseButtonEventArgsDescriptionMetadata.d = function (a) {
        PlotAreaMouseButtonEventArgsDescriptionMetadata.b(a);
        a.ae("PlotAreaMouseButtonEventArgs", function () { return new PlotAreaMouseButtonEventArgsDescription(); });
        a.ad("PlotAreaMouseButtonEventArgs", PlotAreaMouseButtonEventArgsDescriptionMetadata.a);
    };
    PlotAreaMouseButtonEventArgsDescriptionMetadata.$t = markType(PlotAreaMouseButtonEventArgsDescriptionMetadata, 'PlotAreaMouseButtonEventArgsDescriptionMetadata');
    PlotAreaMouseButtonEventArgsDescriptionMetadata.a = null;
    return PlotAreaMouseButtonEventArgsDescriptionMetadata;
}(Base));
export { PlotAreaMouseButtonEventArgsDescriptionMetadata };
/**
 * @hidden
 */
var PlotAreaMouseEventArgsDescriptionMetadata = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PlotAreaMouseEventArgsDescriptionMetadata, _super);
    function PlotAreaMouseEventArgsDescriptionMetadata() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PlotAreaMouseEventArgsDescriptionMetadata.b = function (a) {
        if (PlotAreaMouseEventArgsDescriptionMetadata.a == null) {
            PlotAreaMouseEventArgsDescriptionMetadata.a = new Dictionary$2(String_$type, String_$type, 0);
            PlotAreaMouseEventArgsDescriptionMetadata.c(PlotAreaMouseEventArgsDescriptionMetadata.a);
        }
        if (a.k(PlotAreaMouseEventArgsDescriptionMetadata.a)) {
            return;
        }
        a.ac(PlotAreaMouseEventArgsDescriptionMetadata.a);
        SeriesViewerDescriptionMetadata.d(a);
    };
    PlotAreaMouseEventArgsDescriptionMetadata.c = function (a) {
        a.item("ChartPosition", "Point");
        a.item("PlotAreaPosition", "Point");
        a.item("IsDuringManipulation", "Boolean");
        a.item("Viewer", "ExportedType:SeriesViewer");
    };
    PlotAreaMouseEventArgsDescriptionMetadata.d = function (a) {
        PlotAreaMouseEventArgsDescriptionMetadata.b(a);
        a.ae("PlotAreaMouseEventArgs", function () { return new PlotAreaMouseEventArgsDescription(); });
        a.ad("PlotAreaMouseEventArgs", PlotAreaMouseEventArgsDescriptionMetadata.a);
    };
    PlotAreaMouseEventArgsDescriptionMetadata.$t = markType(PlotAreaMouseEventArgsDescriptionMetadata, 'PlotAreaMouseEventArgsDescriptionMetadata');
    PlotAreaMouseEventArgsDescriptionMetadata.a = null;
    return PlotAreaMouseEventArgsDescriptionMetadata;
}(Base));
export { PlotAreaMouseEventArgsDescriptionMetadata };
/**
 * @hidden
 */
var ChartCursorEventArgsDescriptionMetadata = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ChartCursorEventArgsDescriptionMetadata, _super);
    function ChartCursorEventArgsDescriptionMetadata() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChartCursorEventArgsDescriptionMetadata.b = function (a) {
        if (ChartCursorEventArgsDescriptionMetadata.a == null) {
            ChartCursorEventArgsDescriptionMetadata.a = new Dictionary$2(String_$type, String_$type, 0);
            ChartCursorEventArgsDescriptionMetadata.c(ChartCursorEventArgsDescriptionMetadata.a);
        }
        if (a.k(ChartCursorEventArgsDescriptionMetadata.a)) {
            return;
        }
        a.ac(ChartCursorEventArgsDescriptionMetadata.a);
        SeriesDescriptionMetadata.d(a);
        SeriesViewerDescriptionMetadata.d(a);
    };
    ChartCursorEventArgsDescriptionMetadata.c = function (a) {
        a.item("ItemRef", "(w:Item,wf:Item)DataRef");
        a.item("Series", "ExportedType:Series");
        a.item("SeriesViewer", "ExportedType:SeriesViewer");
    };
    ChartCursorEventArgsDescriptionMetadata.d = function (a) {
        ChartCursorEventArgsDescriptionMetadata.b(a);
        a.ae("ChartCursorEventArgs", function () { return new ChartCursorEventArgsDescription(); });
        a.ad("ChartCursorEventArgs", ChartCursorEventArgsDescriptionMetadata.a);
    };
    ChartCursorEventArgsDescriptionMetadata.$t = markType(ChartCursorEventArgsDescriptionMetadata, 'ChartCursorEventArgsDescriptionMetadata');
    ChartCursorEventArgsDescriptionMetadata.a = null;
    return ChartCursorEventArgsDescriptionMetadata;
}(Base));
export { ChartCursorEventArgsDescriptionMetadata };
/**
 * @hidden
 */
var ChartMouseEventArgsDescriptionMetadata = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ChartMouseEventArgsDescriptionMetadata, _super);
    function ChartMouseEventArgsDescriptionMetadata() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChartMouseEventArgsDescriptionMetadata.b = function (a) {
        if (ChartMouseEventArgsDescriptionMetadata.a == null) {
            ChartMouseEventArgsDescriptionMetadata.a = new Dictionary$2(String_$type, String_$type, 0);
            ChartMouseEventArgsDescriptionMetadata.c(ChartMouseEventArgsDescriptionMetadata.a);
        }
        if (a.k(ChartMouseEventArgsDescriptionMetadata.a)) {
            return;
        }
        a.ac(ChartMouseEventArgsDescriptionMetadata.a);
        SeriesDescriptionMetadata.d(a);
        SeriesViewerDescriptionMetadata.d(a);
    };
    ChartMouseEventArgsDescriptionMetadata.c = function (a) {
        a.item("ItemRef", "(w:Item,wf:Item)DataRef");
        a.item("Series", "ExportedType:Series");
        a.item("PlotAreaPosition", "Point");
        a.item("Chart", "ExportedType:SeriesViewer");
    };
    ChartMouseEventArgsDescriptionMetadata.d = function (a) {
        ChartMouseEventArgsDescriptionMetadata.b(a);
        a.ae("ChartMouseEventArgs", function () { return new ChartMouseEventArgsDescription(); });
        a.ad("ChartMouseEventArgs", ChartMouseEventArgsDescriptionMetadata.a);
    };
    ChartMouseEventArgsDescriptionMetadata.$t = markType(ChartMouseEventArgsDescriptionMetadata, 'ChartMouseEventArgsDescriptionMetadata');
    ChartMouseEventArgsDescriptionMetadata.a = null;
    return ChartMouseEventArgsDescriptionMetadata;
}(Base));
export { ChartMouseEventArgsDescriptionMetadata };
/**
 * @hidden
 */
var DataChartMouseButtonEventArgsDescriptionMetadata = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataChartMouseButtonEventArgsDescriptionMetadata, _super);
    function DataChartMouseButtonEventArgsDescriptionMetadata() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DataChartMouseButtonEventArgsDescriptionMetadata.b = function (a) {
        if (DataChartMouseButtonEventArgsDescriptionMetadata.a == null) {
            DataChartMouseButtonEventArgsDescriptionMetadata.a = new Dictionary$2(String_$type, String_$type, 0);
            DataChartMouseButtonEventArgsDescriptionMetadata.c(DataChartMouseButtonEventArgsDescriptionMetadata.a);
        }
        if (a.k(DataChartMouseButtonEventArgsDescriptionMetadata.a)) {
            return;
        }
        a.ac(DataChartMouseButtonEventArgsDescriptionMetadata.a);
        SeriesDescriptionMetadata.d(a);
        SeriesViewerDescriptionMetadata.d(a);
    };
    DataChartMouseButtonEventArgsDescriptionMetadata.c = function (a) {
        a.item("Handled", "Boolean");
        a.item("CancelSelection", "Boolean");
        a.item("ItemRef", "(w:Item,wf:Item)DataRef");
        a.item("Series", "ExportedType:Series");
        a.item("PlotAreaPosition", "Point");
        a.item("Chart", "ExportedType:SeriesViewer");
    };
    DataChartMouseButtonEventArgsDescriptionMetadata.d = function (a) {
        DataChartMouseButtonEventArgsDescriptionMetadata.b(a);
        a.ae("DataChartMouseButtonEventArgs", function () { return new DataChartMouseButtonEventArgsDescription(); });
        a.ad("DataChartMouseButtonEventArgs", DataChartMouseButtonEventArgsDescriptionMetadata.a);
    };
    DataChartMouseButtonEventArgsDescriptionMetadata.$t = markType(DataChartMouseButtonEventArgsDescriptionMetadata, 'DataChartMouseButtonEventArgsDescriptionMetadata');
    DataChartMouseButtonEventArgsDescriptionMetadata.a = null;
    return DataChartMouseButtonEventArgsDescriptionMetadata;
}(Base));
export { DataChartMouseButtonEventArgsDescriptionMetadata };
/**
 * @hidden
 */
var SeriesViewerDescriptionMetadata = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SeriesViewerDescriptionMetadata, _super);
    function SeriesViewerDescriptionMetadata() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SeriesViewerDescriptionMetadata.b = function (a) {
        if (SeriesViewerDescriptionMetadata.a == null) {
            SeriesViewerDescriptionMetadata.a = new Dictionary$2(String_$type, String_$type, 0);
            SeriesViewerDescriptionMetadata.c(SeriesViewerDescriptionMetadata.a);
        }
        if (a.k(SeriesViewerDescriptionMetadata.a)) {
            return;
        }
        a.ac(SeriesViewerDescriptionMetadata.a);
        PlotAreaMouseButtonEventArgsDescriptionMetadata.d(a);
        PlotAreaMouseEventArgsDescriptionMetadata.d(a);
        ChartCursorEventArgsDescriptionMetadata.d(a);
        DataChartMouseButtonEventArgsDescriptionMetadata.d(a);
        ChartMouseEventArgsDescriptionMetadata.d(a);
        SeriesViewerManipulationEventArgsDescriptionMetadata.d(a);
        RectChangedEventArgsDescriptionMetadata.d(a);
        SeriesViewerSelectedSeriesItemsChangedEventArgsDescriptionMetadata.d(a);
        RefreshCompletedEventArgsDescriptionMetadata.d(a);
        ImageCapturedEventArgsDescriptionMetadata.d(a);
    };
    SeriesViewerDescriptionMetadata.c = function (a) {
        a.item("Series", "Collection:Series:SeriesCollection:Series");
        a.item("Brushes", "BrushCollection:string");
        a.item("Outlines", "BrushCollection:string");
        a.item("MarkerBrushes", "BrushCollection:string");
        a.item("MarkerOutlines", "BrushCollection:string");
        a.item("MarkerAutomaticBehavior", "ExportedType:string:MarkerAutomaticBehavior");
        a.item("MarkerAutomaticBehavior@names", "CircleSmart;None;SmartIndexed;Indexed;Circle;Triangle;Pyramid;Square;Diamond;Pentagon;Hexagon;Tetragram;Pentagram;Hexagram");
        a.item("CrosshairPoint", "Point");
        a.item("LegendRef", "(w:Legend,wf:Legend)DataRef::object");
        a.item("AnimateSeriesWhenAxisRangeChanges", "Boolean");
        a.item("SeriesValueLayerUseLegend", "Boolean");
        a.item("IsWindowSyncedToVisibleRange", "Boolean");
        a.item("ShouldSimulateHoverMoveCrosshairPoint", "Boolean");
        a.item("HighlightingMode", "ExportedType:string:SeriesHighlightingMode");
        a.item("HighlightingMode@names", "Auto;None;BrightenSpecific;Brighten;FadeOthersSpecific;FadeOthers");
        a.item("SelectionMode", "ExportedType:string:SeriesSelectionMode");
        a.item("SelectionMode@names", "Auto;None;Brighten;FadeOthers;GrayscaleOthers;FocusColorThickOutline;FocusColorOutline;SelectionColorThickOutline;SelectionColorOutline;FocusColorFill;SelectionColorFill;ThickOutline");
        a.item("SelectionBehavior", "ExportedType:string:SeriesSelectionBehavior");
        a.item("SelectionBehavior@names", "Auto;PerDataItemMultiSelect;PerDataItemSingleSelect;PerSeriesAndDataItemMultiSelect;PerSeriesAndDataItemSingleSelect;PerSeriesAndDataItemGlobalSingleSelect;PerSeriesMultiSelect;PerSeriesSingleSelect");
        a.item("FocusMode", "ExportedType:string:SeriesSelectionMode");
        a.item("FocusMode@names", "Auto;None;Brighten;FadeOthers;GrayscaleOthers;FocusColorThickOutline;FocusColorOutline;SelectionColorThickOutline;SelectionColorOutline;FocusColorFill;SelectionColorFill;ThickOutline");
        a.item("HorizontalViewScrollbarMode", "ExportedType:string:SeriesViewerScrollbarMode");
        a.item("HorizontalViewScrollbarMode@names", "None;Fading;FadeToLine;Persistent");
        a.item("VerticalViewScrollbarMode", "ExportedType:string:SeriesViewerScrollbarMode");
        a.item("VerticalViewScrollbarMode@names", "None;Fading;FadeToLine;Persistent");
        a.item("HorizontalViewScrollbarPosition", "ExportedType:string:SeriesViewerHorizontalScrollbarPosition");
        a.item("HorizontalViewScrollbarPosition@names", "Auto;InsideBottom;InsideTop");
        a.item("VerticalViewScrollbarPosition", "ExportedType:string:SeriesViewerVerticalScrollbarPosition");
        a.item("VerticalViewScrollbarPosition@names", "Auto;InsideLeft;InsideRight");
        a.item("HorizontalViewScrollbarFill", "Brush");
        a.item("HorizontalViewScrollbarOutline", "Brush");
        a.item("HorizontalViewScrollbarStrokeThickness", "Number:double");
        a.item("HorizontalViewScrollbarMaxOpacity", "Number:double");
        a.item("HorizontalViewScrollbarCornerRadius", "Number:double");
        a.item("HorizontalViewScrollbarHeight", "Number:double");
        a.item("HorizontalViewScrollbarInset", "Number:double");
        a.item("HorizontalViewScrollbarTrackStartInset", "Number:double");
        a.item("HorizontalViewScrollbarTrackEndInset", "Number:double");
        a.item("HorizontalViewScrollbarShouldAddAutoTrackInsets", "Boolean");
        a.item("VerticalViewScrollbarFill", "Brush");
        a.item("VerticalViewScrollbarOutline", "Brush");
        a.item("VerticalViewScrollbarStrokeThickness", "Number:double");
        a.item("VerticalViewScrollbarMaxOpacity", "Number:double");
        a.item("VerticalViewScrollbarCornerRadius", "Number:double");
        a.item("VerticalViewScrollbarWidth", "Number:double");
        a.item("VerticalViewScrollbarInset", "Number:double");
        a.item("VerticalViewScrollbarTrackStartInset", "Number:double");
        a.item("VerticalViewScrollbarTrackEndInset", "Number:double");
        a.item("VerticalViewScrollbarShouldAddAutoTrackInsets", "Boolean");
        a.item("HighlightingBehavior", "ExportedType:string:SeriesHighlightingBehavior");
        a.item("HighlightingBehavior@names", "Auto;DirectlyOver;NearestItems;NearestItemsRetainMainShapes;NearestItemsAndSeries");
        a.item("LegendHighlightingMode", "ExportedType:string:LegendHighlightingMode");
        a.item("LegendHighlightingMode@names", "Auto;None;MatchSeries");
        a.item("LegendItemBadgeShape", "ExportedType:string:LegendItemBadgeShape");
        a.item("LegendItemBadgeShape@names", "Automatic;Circle;Line;Square;Marker;Bar;Column;Hidden");
        a.item("LegendItemBadgeMode", "ExportedType:string:LegendItemBadgeMode");
        a.item("LegendItemBadgeMode@names", "Simplified;MatchSeries");
        a.item("EffectiveViewport", "Rect");
        a.item("WindowRect", "Rect");
        a.item("UseTiledZooming", "Boolean");
        a.item("PreferHigherResolutionTiles", "Boolean");
        a.item("ZoomTileCacheSize", "Number:int");
        a.item("HighlightingTransitionDuration", "Number:int");
        a.item("SelectionTransitionDuration", "Number:int");
        a.item("FocusTransitionDuration", "Number:int");
        a.item("ScrollbarsAnimationDuration", "Number:int");
        a.item("IsPagePanningAllowed", "Boolean");
        a.item("ContentHitTestMode", "ExportedType:string:ChartHitTestMode");
        a.item("ContentHitTestMode@names", "Auto;Computational;ColorEncoded;Mixed;MixedFavoringComputational");
        a.item("ActualContentHitTestMode", "ExportedType:string:ChartHitTestMode");
        a.item("ActualContentHitTestMode@names", "Auto;Computational;ColorEncoded;Mixed;MixedFavoringComputational");
        a.item("IsSurfaceInteractionDisabled", "Boolean");
        a.item("WindowResponse", "ExportedType:string:WindowResponse");
        a.item("WindowResponse@names", "Deferred;Immediate");
        a.item("WindowRectMinWidth", "Number:double");
        a.item("WindowRectMinHeight", "Number:double");
        a.item("ActualWindowRectMinWidth", "Number:double");
        a.item("ActualWindowRectMinHeight", "Number:double");
        a.item("WindowSizeMinWidth", "Number:double");
        a.item("WindowSizeMinHeight", "Number:double");
        a.item("SyncChannel", "String");
        a.item("CrosshairVisibility", "ExportedType:string:Visibility");
        a.item("CrosshairVisibility@names", "Visible;Collapsed");
        a.item("HorizontalCrosshairBrush", "Brush");
        a.item("VerticalCrosshairBrush", "Brush");
        a.item("ZoomCoercionMode", "ExportedType:string:ZoomCoercionMode");
        a.item("ZoomCoercionMode@names", "Auto;AxisConstrained;Unconstrained");
        a.item("PlotAreaBackground", "Brush");
        a.item("ShouldMatchZOrderToSeriesOrder", "Boolean");
        a.item("DefaultInteraction", "ExportedType:string:InteractionState");
        a.item("DefaultInteraction@names", "None;DragZoom;DragPan;DragSelect;Auto");
        a.item("InteractionOverride", "ExportedType:string:InteractionState");
        a.item("InteractionOverride@names", "None;DragZoom;DragPan;DragSelect;Auto");
        a.item("RightButtonDefaultInteraction", "ExportedType:string:InteractionState");
        a.item("RightButtonDefaultInteraction@names", "None;DragZoom;DragPan;DragSelect;Auto");
        a.item("DragModifier", "ExportedType:string:ModifierKeys");
        a.item("DragModifier@names", "None;Alt;Control;Shift;Windows;Apple");
        a.item("PanModifier", "ExportedType:string:ModifierKeys");
        a.item("PanModifier@names", "None;Alt;Control;Shift;Windows;Apple");
        a.item("SelectionModifier", "ExportedType:string:ModifierKeys");
        a.item("SelectionModifier@names", "None;Alt;Control;Shift;Windows;Apple");
        a.item("PreviewRect", "Rect");
        a.item("HighlightedValuesDisplayMode", "ExportedType:string:SeriesHighlightedValuesDisplayMode");
        a.item("HighlightedValuesDisplayMode@names", "Auto;Hidden;Overlay");
        a.item("SelectedSeriesItems", "Collection:ChartSelection:ChartSelectedItemCollection:ChartSelection");
        a.item("FocusedSeriesItems", "Collection:ChartSelection:ChartSelectedItemCollection:ChartSelection");
        a.item("WindowPositionHorizontal", "Number:double");
        a.item("WindowPositionVertical", "Number:double");
        a.item("SelectionBrush", "Brush");
        a.item("FocusBrush", "Brush");
        a.item("ChartTitle", "(w:Title,xam:Title,wf:Title)String");
        a.item("TitleHorizontalAlignment", "ExportedType:string:HorizontalAlignment");
        a.item("TitleHorizontalAlignment@names", "Left;Center;Right;Stretch");
        a.item("SubtitleHorizontalAlignment", "ExportedType:string:HorizontalAlignment");
        a.item("SubtitleHorizontalAlignment@names", "Left;Center;Right;Stretch");
        a.item("TitleTextStyle", "String");
        a.item("SubtitleTextStyle", "String");
        a.item("TitleTextColor", "Brush");
        a.item("SubtitleTextColor", "Brush");
        a.item("TitleTopMargin", "Number:double");
        a.item("TitleLeftMargin", "Number:double");
        a.item("TitleRightMargin", "Number:double");
        a.item("TitleBottomMargin", "Number:double");
        a.item("SubtitleTopMargin", "Number:double");
        a.item("SubtitleLeftMargin", "Number:double");
        a.item("SubtitleRightMargin", "Number:double");
        a.item("SubtitleBottomMargin", "Number:double");
        a.item("Subtitle", "String");
        a.item("TopMargin", "Number:double");
        a.item("LeftMargin", "Number:double");
        a.item("RightMargin", "Number:double");
        a.item("BottomMargin", "Number:double");
        a.item("AutoMarginWidth", "Number:double");
        a.item("AutoMarginHeight", "Number:double");
        a.item("IsAntiAliasingEnabledDuringInteraction", "Boolean");
        a.item("PixelScalingRatio", "Number:double");
        a.item("InteractionPixelScalingRatio", "Number:double");
        a.item("ActualPixelScalingRatio", "Number:double");
        a.item("ActualInteractionPixelScalingRatio", "Number:double");
        a.item("ShouldPanOnMaximumZoom", "Boolean");
        a.item("ActualWindowRect", "Rect");
        a.item("ActualWindowPositionHorizontal", "Number:double");
        a.item("ActualWindowPositionVertical", "Number:double");
        a.item("PlotAreaMouseLeftButtonDownRef", "EventRef::plotAreaMouseLeftButtonDown");
        a.item("PlotAreaMouseLeftButtonUpRef", "EventRef::plotAreaMouseLeftButtonUp");
        a.item("PlotAreaClickedRef", "EventRef::plotAreaClicked");
        a.item("PlotAreaMouseEnterRef", "EventRef::plotAreaMouseEnter");
        a.item("PlotAreaMouseLeaveRef", "EventRef::plotAreaMouseLeave");
        a.item("PlotAreaMouseOverRef", "EventRef::plotAreaMouseOver");
        a.item("SeriesCursorMouseMoveRef", "EventRef::seriesCursorMouseMove");
        a.item("SeriesMouseLeftButtonDownRef", "EventRef::seriesMouseLeftButtonDown");
        a.item("SeriesMouseLeftButtonUpRef", "EventRef::seriesMouseLeftButtonUp");
        a.item("SeriesClickRef", "EventRef::seriesClick");
        a.item("SeriesMouseMoveRef", "EventRef::seriesMouseMove");
        a.item("SeriesMouseEnterRef", "EventRef::seriesMouseEnter");
        a.item("SeriesMouseLeaveRef", "EventRef::seriesMouseLeave");
        a.item("ViewerManipulationStartingRef", "EventRef::viewerManipulationStarting");
        a.item("ViewerManipulationEndingRef", "EventRef::viewerManipulationEnding");
        a.item("WindowRectChangedRef", "EventRef::windowRectChanged");
        a.item("SizeChangedRef", "EventRef::sizeChanged");
        a.item("ActualWindowRectChangedRef", "EventRef::actualWindowRectChanged");
        a.item("GridAreaRectChangedRef", "EventRef::gridAreaRectChanged");
        a.item("SelectedSeriesItemsChangedRef", "EventRef::selectedSeriesItemsChanged");
        a.item("FocusedSeriesItemsChangedRef", "EventRef::focusedSeriesItemsChanged");
        a.item("RefreshCompletedRef", "EventRef::refreshCompleted");
        a.item("ImageCapturedRef", "EventRef::imageCaptured");
        SeriesViewerDescriptionMetadata.e(a);
    };
    SeriesViewerDescriptionMetadata.d = function (a) {
        SeriesViewerDescriptionMetadata.b(a);
        a.ad("SeriesViewer", SeriesViewerDescriptionMetadata.a);
    };
    SeriesViewerDescriptionMetadata.e = function (a) {
        a.item("Background", "(wf:BackColor/ColorTransform)Brush");
        a.item("TitleFontFamily", "(w:TitleFontFamily/FontFamilyTransform,xam:TitleFontFamily,wf:TitleFontFamily,web:TitleTextStyle/FontFamilyTransform)String");
        a.item("TitleFontSize", "(w:TitleFontSize,xam:TitleFontSize,wf:TitleFontSize,web:TitleTextStyle/FontSizeTransform)Number:double");
        a.item("TitleFontStyle", "(w:TitleFontStyle/FontStyleTransform,xam:TitleTextStyle/FontStyleTransform,wf:TitleTextStyle/FontStyleTransform,web:TitleTextStyle/FontStyleTransform)String");
        a.item("TitleFontWeight", "(w:TitleFontWeight/FontWeightTransform,xam:TitleTextStyle/FontWeightTransform,wf:TitleTextStyle/FontWeightTransform,web:TitleTextStyle/FontWeightTransform)String");
        a.item("SubtitleFontFamily", "(w:SubtitleFontFamily/FontFamilyTransform,xam:SubtitleFontFamily,wf:SubtitleFontFamily,web:SubtitleTextStyle/FontFamilyTransform)String");
        a.item("SubtitleFontSize", "(w:SubtitleFontSize,xam:SubtitleFontSize,wf:SubtitleFontSize,web:SubtitleTextStyle/FontSizeTransform)Number:double");
        a.item("SubtitleFontStyle", "(w:SubtitleFontStyle/FontStyleTransform,xam:SubtitleTextStyle/FontStyleTransform,wf:SubtitleTextStyle/FontStyleTransform,web:SubtitleTextStyle/FontStyleTransform)String");
        a.item("SubtitleFontWeight", "(w:SubtitleFontWeight/FontWeightTransform,xam:SubtitleTextStyle/FontWeightTransform,wf:SubtitleTextStyle/FontWeightTransform,web:SubtitleTextStyle/FontWeightTransform)String");
        a.item("SynchronizeHorizontally", "Boolean");
        a.item("SynchronizeVertically", "Boolean");
    };
    SeriesViewerDescriptionMetadata.$t = markType(SeriesViewerDescriptionMetadata, 'SeriesViewerDescriptionMetadata');
    SeriesViewerDescriptionMetadata.a = null;
    return SeriesViewerDescriptionMetadata;
}(Base));
export { SeriesViewerDescriptionMetadata };
