// igniteui-react-core/esm5/lib/PopupPointerPosition.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Indicates where the popup pointer should be positioned.
 */
export var PopupPointerPosition = /*@__PURE__*/ (function (PopupPointerPosition) {
    PopupPointerPosition[PopupPointerPosition["Auto"] = 0] = "Auto";
    PopupPointerPosition[PopupPointerPosition["Top"] = 1] = "Top";
    PopupPointerPosition[PopupPointerPosition["Bottom"] = 2] = "Bottom";
    PopupPointerPosition[PopupPointerPosition["Left"] = 3] = "Left";
    PopupPointerPosition[PopupPointerPosition["Right"] = 4] = "Right";
    return PopupPointerPosition;
})({});
/**
 * @hidden
 */
export var PopupPointerPosition_$type = /*@__PURE__*/ markEnum('PopupPointerPosition', 'Auto,0|Top,1|Bottom,2|Left,3|Right,4');
