// igniteui-react-core/esm5/lib/PopupDirection.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Indicates which direction the popup will pop up in relation to the exclusion rectangle.
 */
export var PopupDirection = /*@__PURE__*/ (function (PopupDirection) {
    /**
     * No specified direction.
     */
    PopupDirection[PopupDirection["Auto"] = 0] = "Auto";
    /**
     * Popup should be displayed downward.
     */
    PopupDirection[PopupDirection["Down"] = 1] = "Down";
    /**
     * Popup should be displayed upward.
     */
    PopupDirection[PopupDirection["Up"] = 2] = "Up";
    /**
     * Popup should be displayed to the left.
     */
    PopupDirection[PopupDirection["Left"] = 3] = "Left";
    /**
     * Popup should be displayed to the right.
     */
    PopupDirection[PopupDirection["Right"] = 4] = "Right";
    return PopupDirection;
})({});
/**
 * @hidden
 */
export var PopupDirection_$type = /*@__PURE__*/ markEnum('PopupDirection', 'Auto,0|Down,1|Up,2|Left,3|Right,4');
