// igniteui-react-core/esm5/lib/ProportionalCategoryAngleAxisDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { CategoryAngleAxisDescription } from "./CategoryAngleAxisDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ProportionalCategoryAngleAxisDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ProportionalCategoryAngleAxisDescription, _super);
    function ProportionalCategoryAngleAxisDescription() {
        var _this = _super.call(this) || this;
        _this.el = null;
        _this.eh = 0;
        _this.ek = null;
        _this.ej = null;
        return _this;
    }
    ProportionalCategoryAngleAxisDescription.prototype.get_type = function () {
        return "ProportionalCategoryAngleAxis";
    };
    Object.defineProperty(ProportionalCategoryAngleAxisDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            this.el = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ProportionalCategoryAngleAxisDescription.prototype, "othersCategoryThreshold", {
        get: function () {
            return this.eh;
        },
        set: function (a) {
            this.eh = a;
            this.g("OthersCategoryThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ProportionalCategoryAngleAxisDescription.prototype, "othersCategoryType", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.g("OthersCategoryType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ProportionalCategoryAngleAxisDescription.prototype, "othersCategoryText", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("OthersCategoryText");
        },
        enumerable: false,
        configurable: true
    });
    ProportionalCategoryAngleAxisDescription.$t = markType(ProportionalCategoryAngleAxisDescription, 'ProportionalCategoryAngleAxisDescription', CategoryAngleAxisDescription.$);
    return ProportionalCategoryAngleAxisDescription;
}(CategoryAngleAxisDescription));
export { ProportionalCategoryAngleAxisDescription };
