// igniteui-react-core/esm5/lib/RadialBaseChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DomainChartDescription } from "./DomainChartDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var RadialBaseChartDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RadialBaseChartDescription, _super);
    function RadialBaseChartDescription() {
        var _this = _super.call(this) || this;
        _this.t5 = null;
        _this.un = null;
        _this.sa = 0;
        _this.sc = 0;
        _this.sb = 0;
        _this.r9 = 0;
        _this.sr = 0;
        _this.st = 0;
        _this.ss = 0;
        _this.sq = 0;
        _this.ua = null;
        _this.us = null;
        _this.t3 = null;
        _this.t4 = null;
        _this.sl = 0;
        _this.s2 = 0;
        _this.sk = 0;
        _this.s1 = 0;
        _this.sn = 0;
        _this.s4 = 0;
        _this.sm = 0;
        _this.s3 = 0;
        _this.sj = 0;
        _this.s0 = 0;
        _this.ul = null;
        _this.u3 = null;
        _this.ub = null;
        _this.ut = null;
        _this.um = null;
        _this.u4 = null;
        _this.t9 = null;
        _this.ur = null;
        _this.ue = null;
        _this.uw = null;
        _this.sd = 0;
        _this.su = 0;
        _this.se = 0;
        _this.sv = 0;
        _this.ug = null;
        _this.uy = null;
        _this.uh = null;
        _this.uz = null;
        _this.sf = 0;
        _this.sw = 0;
        _this.sg = 0;
        _this.sx = 0;
        _this.ui = null;
        _this.u0 = null;
        _this.sh = 0;
        _this.sy = 0;
        _this.uj = null;
        _this.u1 = null;
        _this.uf = null;
        _this.ux = null;
        _this.r8 = 0;
        _this.sp = 0;
        _this.r7 = 0;
        _this.so = 0;
        _this.si = 0;
        _this.sz = 0;
        _this.r3 = false;
        _this.r4 = false;
        _this.uk = null;
        _this.u2 = null;
        _this.t7 = null;
        _this.up = null;
        _this.uc = null;
        _this.uu = null;
        _this.ud = null;
        _this.uv = null;
        _this.uq = null;
        _this.t8 = null;
        _this.t6 = null;
        _this.rz = null;
        _this.uo = null;
        _this.r0 = null;
        return _this;
    }
    RadialBaseChartDescription.prototype.get_type = function () {
        return "RadialBaseChart";
    };
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisFormatLabelRef", {
        get: function () {
            return this.t5;
        },
        set: function (a) {
            this.t5 = a;
            this.g("AngleAxisFormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisFormatLabelRef", {
        get: function () {
            return this.un;
        },
        set: function (a) {
            this.un = a;
            this.g("ValueAxisFormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelLeftMargin", {
        get: function () {
            return this.sa;
        },
        set: function (a) {
            this.sa = a;
            this.g("AngleAxisLabelLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelTopMargin", {
        get: function () {
            return this.sc;
        },
        set: function (a) {
            this.sc = a;
            this.g("AngleAxisLabelTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelRightMargin", {
        get: function () {
            return this.sb;
        },
        set: function (a) {
            this.sb = a;
            this.g("AngleAxisLabelRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelBottomMargin", {
        get: function () {
            return this.r9;
        },
        set: function (a) {
            this.r9 = a;
            this.g("AngleAxisLabelBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelLeftMargin", {
        get: function () {
            return this.sr;
        },
        set: function (a) {
            this.sr = a;
            this.g("ValueAxisLabelLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelTopMargin", {
        get: function () {
            return this.st;
        },
        set: function (a) {
            this.st = a;
            this.g("ValueAxisLabelTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelRightMargin", {
        get: function () {
            return this.ss;
        },
        set: function (a) {
            this.ss = a;
            this.g("ValueAxisLabelRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelBottomMargin", {
        get: function () {
            return this.sq;
        },
        set: function (a) {
            this.sq = a;
            this.g("ValueAxisLabelBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelTextColor", {
        get: function () {
            return this.ua;
        },
        set: function (a) {
            this.ua = a;
            this.g("AngleAxisLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelTextColor", {
        get: function () {
            return this.us;
        },
        set: function (a) {
            this.us = a;
            this.g("ValueAxisLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "actualAngleAxisLabelTextColor", {
        get: function () {
            return this.t3;
        },
        set: function (a) {
            this.t3 = a;
            this.g("ActualAngleAxisLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "actualValueAxisLabelTextColor", {
        get: function () {
            return this.t4;
        },
        set: function (a) {
            this.t4 = a;
            this.g("ActualValueAxisLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTitleMargin", {
        get: function () {
            return this.sl;
        },
        set: function (a) {
            this.sl = a;
            this.g("AngleAxisTitleMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTitleMargin", {
        get: function () {
            return this.s2;
        },
        set: function (a) {
            this.s2 = a;
            this.g("ValueAxisTitleMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTitleLeftMargin", {
        get: function () {
            return this.sk;
        },
        set: function (a) {
            this.sk = a;
            this.g("AngleAxisTitleLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTitleLeftMargin", {
        get: function () {
            return this.s1;
        },
        set: function (a) {
            this.s1 = a;
            this.g("ValueAxisTitleLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTitleTopMargin", {
        get: function () {
            return this.sn;
        },
        set: function (a) {
            this.sn = a;
            this.g("AngleAxisTitleTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTitleTopMargin", {
        get: function () {
            return this.s4;
        },
        set: function (a) {
            this.s4 = a;
            this.g("ValueAxisTitleTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTitleRightMargin", {
        get: function () {
            return this.sm;
        },
        set: function (a) {
            this.sm = a;
            this.g("AngleAxisTitleRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTitleRightMargin", {
        get: function () {
            return this.s3;
        },
        set: function (a) {
            this.s3 = a;
            this.g("ValueAxisTitleRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTitleBottomMargin", {
        get: function () {
            return this.sj;
        },
        set: function (a) {
            this.sj = a;
            this.g("AngleAxisTitleBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTitleBottomMargin", {
        get: function () {
            return this.s0;
        },
        set: function (a) {
            this.s0 = a;
            this.g("ValueAxisTitleBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTitleTextColor", {
        get: function () {
            return this.ul;
        },
        set: function (a) {
            this.ul = a;
            this.g("AngleAxisTitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTitleTextColor", {
        get: function () {
            return this.u3;
        },
        set: function (a) {
            this.u3 = a;
            this.g("ValueAxisTitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelTextStyle", {
        get: function () {
            return this.ub;
        },
        set: function (a) {
            this.ub = a;
            this.g("AngleAxisLabelTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelTextStyle", {
        get: function () {
            return this.ut;
        },
        set: function (a) {
            this.ut = a;
            this.g("ValueAxisLabelTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTitleTextStyle", {
        get: function () {
            return this.um;
        },
        set: function (a) {
            this.um = a;
            this.g("AngleAxisTitleTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTitleTextStyle", {
        get: function () {
            return this.u4;
        },
        set: function (a) {
            this.u4 = a;
            this.g("ValueAxisTitleTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelRef", {
        get: function () {
            return this.t9;
        },
        set: function (a) {
            this.t9 = a;
            this.g("AngleAxisLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelRef", {
        get: function () {
            return this.ur;
        },
        set: function (a) {
            this.ur = a;
            this.g("ValueAxisLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisMajorStroke", {
        get: function () {
            return this.ue;
        },
        set: function (a) {
            this.ue = a;
            this.g("AngleAxisMajorStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisMajorStroke", {
        get: function () {
            return this.uw;
        },
        set: function (a) {
            this.uw = a;
            this.g("ValueAxisMajorStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisMajorStrokeThickness", {
        get: function () {
            return this.sd;
        },
        set: function (a) {
            this.sd = a;
            this.g("AngleAxisMajorStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisMajorStrokeThickness", {
        get: function () {
            return this.su;
        },
        set: function (a) {
            this.su = a;
            this.g("ValueAxisMajorStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisMinorStrokeThickness", {
        get: function () {
            return this.se;
        },
        set: function (a) {
            this.se = a;
            this.g("AngleAxisMinorStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisMinorStrokeThickness", {
        get: function () {
            return this.sv;
        },
        set: function (a) {
            this.sv = a;
            this.g("ValueAxisMinorStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisStrip", {
        get: function () {
            return this.ug;
        },
        set: function (a) {
            this.ug = a;
            this.g("AngleAxisStrip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisStrip", {
        get: function () {
            return this.uy;
        },
        set: function (a) {
            this.uy = a;
            this.g("ValueAxisStrip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisStroke", {
        get: function () {
            return this.uh;
        },
        set: function (a) {
            this.uh = a;
            this.g("AngleAxisStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisStroke", {
        get: function () {
            return this.uz;
        },
        set: function (a) {
            this.uz = a;
            this.g("ValueAxisStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisStrokeThickness", {
        get: function () {
            return this.sf;
        },
        set: function (a) {
            this.sf = a;
            this.g("AngleAxisStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisStrokeThickness", {
        get: function () {
            return this.sw;
        },
        set: function (a) {
            this.sw = a;
            this.g("ValueAxisStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTickLength", {
        get: function () {
            return this.sg;
        },
        set: function (a) {
            this.sg = a;
            this.g("AngleAxisTickLength");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTickLength", {
        get: function () {
            return this.sx;
        },
        set: function (a) {
            this.sx = a;
            this.g("ValueAxisTickLength");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTickStroke", {
        get: function () {
            return this.ui;
        },
        set: function (a) {
            this.ui = a;
            this.g("AngleAxisTickStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTickStroke", {
        get: function () {
            return this.u0;
        },
        set: function (a) {
            this.u0 = a;
            this.g("ValueAxisTickStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTickStrokeThickness", {
        get: function () {
            return this.sh;
        },
        set: function (a) {
            this.sh = a;
            this.g("AngleAxisTickStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTickStrokeThickness", {
        get: function () {
            return this.sy;
        },
        set: function (a) {
            this.sy = a;
            this.g("ValueAxisTickStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTitle", {
        get: function () {
            return this.uj;
        },
        set: function (a) {
            this.uj = a;
            this.g("AngleAxisTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTitle", {
        get: function () {
            return this.u1;
        },
        set: function (a) {
            this.u1 = a;
            this.g("ValueAxisTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisMinorStroke", {
        get: function () {
            return this.uf;
        },
        set: function (a) {
            this.uf = a;
            this.g("AngleAxisMinorStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisMinorStroke", {
        get: function () {
            return this.ux;
        },
        set: function (a) {
            this.ux = a;
            this.g("ValueAxisMinorStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelAngle", {
        get: function () {
            return this.r8;
        },
        set: function (a) {
            this.r8 = a;
            this.g("AngleAxisLabelAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelAngle", {
        get: function () {
            return this.sp;
        },
        set: function (a) {
            this.sp = a;
            this.g("ValueAxisLabelAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisExtent", {
        get: function () {
            return this.r7;
        },
        set: function (a) {
            this.r7 = a;
            this.g("AngleAxisExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisExtent", {
        get: function () {
            return this.so;
        },
        set: function (a) {
            this.so = a;
            this.g("ValueAxisExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTitleAngle", {
        get: function () {
            return this.si;
        },
        set: function (a) {
            this.si = a;
            this.g("AngleAxisTitleAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTitleAngle", {
        get: function () {
            return this.sz;
        },
        set: function (a) {
            this.sz = a;
            this.g("ValueAxisTitleAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisInverted", {
        get: function () {
            return this.r3;
        },
        set: function (a) {
            this.r3 = a;
            this.g("AngleAxisInverted");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisInverted", {
        get: function () {
            return this.r4;
        },
        set: function (a) {
            this.r4 = a;
            this.g("ValueAxisInverted");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisTitleAlignment", {
        get: function () {
            return this.uk;
        },
        set: function (a) {
            this.uk = a;
            this.g("AngleAxisTitleAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisTitleAlignment", {
        get: function () {
            return this.u2;
        },
        set: function (a) {
            this.u2 = a;
            this.g("ValueAxisTitleAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelHorizontalAlignment", {
        get: function () {
            return this.t7;
        },
        set: function (a) {
            this.t7 = a;
            this.g("AngleAxisLabelHorizontalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelHorizontalAlignment", {
        get: function () {
            return this.up;
        },
        set: function (a) {
            this.up = a;
            this.g("ValueAxisLabelHorizontalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelVerticalAlignment", {
        get: function () {
            return this.uc;
        },
        set: function (a) {
            this.uc = a;
            this.g("AngleAxisLabelVerticalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelVerticalAlignment", {
        get: function () {
            return this.uu;
        },
        set: function (a) {
            this.uu = a;
            this.g("ValueAxisLabelVerticalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelVisibility", {
        get: function () {
            return this.ud;
        },
        set: function (a) {
            this.ud = a;
            this.g("AngleAxisLabelVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelVisibility", {
        get: function () {
            return this.uv;
        },
        set: function (a) {
            this.uv = a;
            this.g("ValueAxisLabelVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelLocation", {
        get: function () {
            return this.uq;
        },
        set: function (a) {
            this.uq = a;
            this.g("ValueAxisLabelLocation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelLocation", {
        get: function () {
            return this.t8;
        },
        set: function (a) {
            this.t8 = a;
            this.g("AngleAxisLabelLocation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelFormat", {
        get: function () {
            return this.t6;
        },
        set: function (a) {
            this.t6 = a;
            this.g("AngleAxisLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "angleAxisLabelFormatSpecifiers", {
        get: function () {
            return this.rz;
        },
        set: function (a) {
            this.rz = a;
            this.g("AngleAxisLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelFormat", {
        get: function () {
            return this.uo;
        },
        set: function (a) {
            this.uo = a;
            this.g("ValueAxisLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseChartDescription.prototype, "valueAxisLabelFormatSpecifiers", {
        get: function () {
            return this.r0;
        },
        set: function (a) {
            this.r0 = a;
            this.g("ValueAxisLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    RadialBaseChartDescription.$t = markType(RadialBaseChartDescription, 'RadialBaseChartDescription', DomainChartDescription.$);
    return RadialBaseChartDescription;
}(DomainChartDescription));
export { RadialBaseChartDescription };
