// igniteui-react-core/esm5/lib/RadialPieSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnchoredRadialSeriesDescription } from "./AnchoredRadialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var RadialPieSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RadialPieSeriesDescription, _super);
    function RadialPieSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.ht = 0;
        _this.hu = 0;
        _this.hy = null;
        _this.hx = null;
        return _this;
    }
    RadialPieSeriesDescription.prototype.get_type = function () {
        return "RadialPieSeries";
    };
    Object.defineProperty(RadialPieSeriesDescription.prototype, "radiusX", {
        get: function () {
            return this.ht;
        },
        set: function (a) {
            this.ht = a;
            this.g("RadiusX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialPieSeriesDescription.prototype, "radiusY", {
        get: function () {
            return this.hu;
        },
        set: function (a) {
            this.hu = a;
            this.g("RadiusY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialPieSeriesDescription.prototype, "legendLabelMemberPath", {
        get: function () {
            return this.hy;
        },
        set: function (a) {
            this.hy = a;
            this.g("LegendLabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialPieSeriesDescription.prototype, "legendEmptyValuesMode", {
        get: function () {
            return this.hx;
        },
        set: function (a) {
            this.hx = a;
            this.g("LegendEmptyValuesMode");
        },
        enumerable: false,
        configurable: true
    });
    RadialPieSeriesDescription.$t = markType(RadialPieSeriesDescription, 'RadialPieSeriesDescription', AnchoredRadialSeriesDescription.$);
    return RadialPieSeriesDescription;
}(AnchoredRadialSeriesDescription));
export { RadialPieSeriesDescription };
