// igniteui-react-core/esm5/lib/RangeCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var RangeCategorySeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RangeCategorySeriesDescription, _super);
    function RangeCategorySeriesDescription() {
        var _this = _super.call(this) || this;
        _this.ga = null;
        _this.f9 = null;
        _this.f8 = null;
        _this.f7 = null;
        return _this;
    }
    RangeCategorySeriesDescription.prototype.get_type = function () {
        return "RangeCategorySeries";
    };
    Object.defineProperty(RangeCategorySeriesDescription.prototype, "lowMemberPath", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("LowMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RangeCategorySeriesDescription.prototype, "highMemberPath", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("HighMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RangeCategorySeriesDescription.prototype, "highlightedLowMemberPath", {
        get: function () {
            return this.f8;
        },
        set: function (a) {
            this.f8 = a;
            this.g("HighlightedLowMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RangeCategorySeriesDescription.prototype, "highlightedHighMemberPath", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("HighlightedHighMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    RangeCategorySeriesDescription.$t = markType(RangeCategorySeriesDescription, 'RangeCategorySeriesDescription', CategorySeriesDescription.$);
    return RangeCategorySeriesDescription;
}(CategorySeriesDescription));
export { RangeCategorySeriesDescription };
