// igniteui-react-core/esm5/lib/RearrangedListView.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, String_$type, Number_$type, typeCast, Array_$type, Boolean_$type, markType } from "./type";
import { ISupportsDataChangeNotifications_$type } from "./ISupportsDataChangeNotifications";
import { List$1 } from "./List$1";
import { SectionData } from "./SectionData";
import { SectionMap } from "./SectionMap";
import { DataSourceRootSummariesChangedEventArgs } from "./DataSourceRootSummariesChangedEventArgs";
import { Dictionary$2 } from "./Dictionary$2";
import { DataSourceSpecialRow } from "./DataSourceSpecialRow";
import { SummaryGroup } from "./SummaryGroup";
import { RemovedShiftedRowInfo } from "./RemovedShiftedRowInfo";
import { dateNow } from "./date";
/**
 * @hidden
 */
var RearrangedListView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RearrangedListView, _super);
    function RearrangedListView(a) {
        var _this = _super.call(this) || this;
        _this.o = null;
        _this.a6 = null;
        _this.aa = new SectionMap();
        _this.e = null;
        _this.a = null;
        _this.b = null;
        _this.c = null;
        _this.d = null;
        _this.ba = 0;
        _this.a9 = 0;
        _this.rootSummariesChanged = null;
        _this.an = false;
        _this.at = false;
        _this.p = null;
        _this.a2 = null;
        _this.a1 = null;
        _this.t = null;
        _this.ab = null;
        _this.ad = null;
        _this.n = null;
        _this.af = null;
        _this.ag = null;
        _this.ah = null;
        _this.aq = false;
        _this.l = 0;
        _this.k = 0;
        _this.aw = false;
        _this.av = false;
        _this.ax = false;
        _this.ay = false;
        _this.ai = false;
        _this.a5 = new List$1(Array_$type, 0);
        _this.a0 = new Dictionary$2(String_$type, Boolean_$type, 0);
        _this.o = a;
        return _this;
    }
    Object.defineProperty(RearrangedListView.prototype, "bm", {
        get: function () {
            return this.a9;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RearrangedListView.prototype, "i", {
        get: function () {
            return this.e;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RearrangedListView.prototype, "a8", {
        get: function () {
            return this.a6;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RearrangedListView.prototype, "ar", {
        get: function () {
            if (!this.n.isSectionContentVisible) {
                return true;
            }
            return this.n.isSectionCollapsable;
        },
        enumerable: false,
        configurable: true
    });
    RearrangedListView.prototype.x = function (a) {
        return this.aa.e(a);
    };
    RearrangedListView.prototype.w = function () {
        return this.aa.d();
    };
    Object.defineProperty(RearrangedListView.prototype, "q", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RearrangedListView.prototype, "a4", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RearrangedListView.prototype, "a3", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RearrangedListView.prototype, "u", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RearrangedListView.prototype, "ac", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RearrangedListView.prototype, "ae", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RearrangedListView.prototype, "m", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RearrangedListView.prototype, "as", {
        get: function () {
            if (!this.n.isSectionContentVisible) {
                return false;
            }
            if (this.ar) {
                return this.n.isSectionExpandedDefault;
            }
            return true;
        },
        enumerable: false,
        configurable: true
    });
    RearrangedListView.prototype.b9 = function () {
        var _this = this;
        if (this.a6 == this.o || this.a6 == null) {
            this.a6 = new List$1(Base.$, 0);
        }
        this.a6.clear();
        if (!this.ar) {
            this.a0.clear();
        }
        if (this.o == null) {
            return;
        }
        var a = this.o.actualCount;
        this.a9 = 0;
        this.ba = this.bc();
        this.bt(null);
        var b = this.ah.k.count > 0 && this.m != 2 && this.m != 3;
        if (this.t == null) {
            for (var c = 0; c < a; c++) {
                var d = this.o.getItemAtIndex(c);
                this.a6.add1(d);
                if (b) {
                    this.bs(null, d);
                }
            }
        }
        else {
            for (var e = 0; e < a; e++) {
                var f = this.o.getItemAtIndex(e);
                if (this.t(f)) {
                    this.a6.add1(f);
                    if (b) {
                        this.bs(null, f);
                    }
                }
            }
        }
        this.bw(null);
        if (this.a2 != null) {
            this.a6.z(this.a2);
        }
        this.ai = false;
        this.aa.v();
        if (this.af != null && this.af.k.count > 0) {
            this.ai = true;
            this.b3();
        }
        else {
            this.aa.t.add(((function () {
                var $ret = new SectionData();
                $ret.ae = 0;
                $ret.e = null;
                $ret.f = null;
                $ret.af = 0;
                $ret.ad = _this.a6.count;
                $ret.c = new Array(0);
                $ret.d = new Array(1);
                return $ret;
            })()));
        }
        if (this.ay && b) {
            this.by(this.e);
        }
        var g = this.q;
        var h = new List$1(Base.$, 0);
        if (this.a5.count > 0 && this.ax) {
            if (!this.ai) {
                var i = this.aa.t._inner[0];
                for (var j = 0; j < this.a5.count; j++) {
                    var k = this.n.indexOfKey(this.a5._inner[j]);
                    if (k != -1) {
                        this.bl(i, this.a6._inner[k], j);
                    }
                }
            }
            else {
                for (var l = 0; l < this.a5.count; l++) {
                    var m = this.n.indexOfKey(this.a5._inner[l]);
                    if (m != -1) {
                        var n = this.aa.e(m);
                        this.bl(n, this.a6._inner[m], l);
                    }
                }
            }
        }
        if (this.ac != null) {
            this.ac.notifyClearItems();
        }
    };
    RearrangedListView.prototype.bn = function (a, b) {
        var c = null;
        if (a.p) {
            if (b) {
                var d = a.af + a.v() + a.ah + a.aa;
                c = this.a6._inner[d];
            }
            else {
                var e = a.ad - a.u - a.ag;
                c = this.a6._inner[e];
            }
        }
        return c;
    };
    Object.defineProperty(RearrangedListView.prototype, "a7", {
        get: function () {
            return this.aa.t;
        },
        enumerable: false,
        configurable: true
    });
    RearrangedListView.prototype.b3 = function () {
        if (this.a6.count == 0) {
            return;
        }
        var a = dateNow();
        var b = this.f();
        var c = this.g();
        var d = this.k == 1 ? c.length : 1;
        if (this.b == null || this.b.length != c.length) {
            this.a = new Array(d);
            this.b = new Array(d);
            for (var e = 0; e < d; e++) {
                this.a[e] = null;
                this.b[e] = null;
            }
        }
        this.c = new Array(d);
        for (var f = 0; f < d; f++) {
            this.c[f] = new Array(this.ah.k.count);
        }
        var g = this.m == 3 || this.m == 1;
        var h = new List$1(Base.$, 0);
        var i = null;
        var j = null;
        if (this.n.isSectionContentVisible || !this.aw) {
            for (var k = 0; k < this.a6.count; k++) {
                var l = this.a6._inner[k];
                if (l == null) {
                    continue;
                }
                if (i == null) {
                    var m = this.y(k, l, c);
                    k += this.be(m, k, b, c, true);
                    this.bu(m, true, g);
                    this.aa.z(m, 1);
                    this.bs(m, l);
                    j = l;
                }
                else {
                    if (this.a3.compare(i, l) != 0) {
                        var n = this.aa.t._inner[this.aa.t.count - 1];
                        if (n.q) {
                            k += this.bd(k, j, b, c);
                        }
                        this.bv(k - 1, j, b);
                        var o = this.y(k, l, c);
                        this.bu(o, false, g);
                        var p = this.be(o, k, b, c, true);
                        k += p;
                        o.q = this.aa.r(o);
                        this.aa.z(o, 1);
                        this.bs(o, l);
                        j = l;
                    }
                    else {
                        var q = this.aa.t._inner[this.aa.t.count - 1];
                        this.aa.z(q, 1);
                        this.bs(q, l);
                    }
                }
                if (k == this.a6.count - 1) {
                    var r = this.aa.t._inner[this.aa.t.count - 1];
                    if (r.q) {
                        k += this.bd(k + 1, j, b, c);
                    }
                    this.bv(k, j, b);
                }
                i = l;
            }
            if (!g) {
                if (this.k == 1) {
                    for (var s = 0; s < c.length; s++) {
                        if (this.b[s] == null) {
                            continue;
                        }
                        for (var t = 0; t < this.ah.k.count; t++) {
                            this.b[s].i[s][t] = this.c[s][t].getResults();
                            this.b[s].i[s][t].summaryIndex = t;
                            this.b[s].i[s][t].groupKey = this.b[s].f;
                        }
                        for (var u = 0; u < this.b[s].h.length; u++) {
                            if (this.b[s].h[u] == -1) {
                                this.b[s].h[u] = this.aa.t._inner[this.aa.t.count - 1].ad + 1;
                            }
                        }
                    }
                }
                else {
                    if (this.b[0] != null) {
                        for (var v = 0; v < this.ah.k.count; v++) {
                            this.b[0].i[0][v] = this.c[0][v].getResults();
                            this.b[0].i[0][v].summaryIndex = v;
                            this.b[0].i[0][v].groupKey = this.b[0].f;
                        }
                        this.b[0].h[0] = this.aa.t._inner[this.aa.t.count - 1].ad + 1;
                    }
                }
            }
            this.bz();
            for (var w = 0; w < this.aa.t.count; w++) {
                var x = this.aa.t._inner[w];
                if (this.aw) {
                    if (!x.q) {
                        var y = x.ad - x.af + 1;
                        if (this.k == 1 && this.af.k.count > 1 && this.an && this.at) {
                            var z = x.w();
                            var aa = false;
                            if (w + 1 < this.aa.t.count && this.aa.t._inner[w + 1].w() == 0) {
                                aa = true;
                            }
                            if (aa) {
                                var ab = x.v();
                                y = ab + x.d[x.d.length - 1];
                            }
                        }
                        this.a6.v(x.af, y);
                        this.aa.ae(x.ae);
                        this.aa.ab(x, -y, -y);
                        w--;
                    }
                    else if (!x.p) {
                        var ac = 0;
                        var ad = x.w();
                        for (var ae = ad; ae < x.c.length; ae++) {
                            ac = x.b[ae];
                            if (!x.c[ae]) {
                                break;
                            }
                        }
                        var af = x.af + ac + 1;
                        var ag = x.d[x.d.length - 1];
                        this.a6.v(af, ag);
                        x.ad -= ag;
                        this.aa.ac(x, -ag, -ag);
                    }
                }
            }
        }
        else {
            for (var ah = 0; ah < this.a6.count; ah++) {
                var ai = this.a6._inner[ah];
                if (ai == null) {
                    continue;
                }
                if (i == null) {
                    var aj = this.y(ah, ai, c);
                    this.bf(h, aj, ai, b, c, true);
                    this.bu(aj, true, g);
                    this.aa.z(aj, 1);
                    this.bs(aj, ai);
                    j = ai;
                }
                else {
                    if (this.a3.compare(i, ai) != 0) {
                        var ak = this.aa.t._inner[this.aa.t.count - 1];
                        this.bv(h.count - 1, j, b);
                        var al = this.y(ah, ai, c);
                        this.bu(al, false, g);
                        this.bf(h, al, ai, b, c, true);
                        al.q = this.aa.r(al);
                        this.aa.z(al, 1);
                        this.bs(al, ai);
                        j = ai;
                    }
                    else {
                        var am = this.aa.t._inner[this.aa.t.count - 1];
                        this.aa.z(am, 1);
                        this.bs(am, ai);
                    }
                }
                if (ah == this.a6.count - 1) {
                    var an = this.aa.t._inner[this.aa.t.count - 1];
                    this.bv(ah, j, b);
                }
                i = ai;
            }
            this.a6 = h;
        }
        for (var ao = 0; ao < this.b.length; ao++) {
            this.a[ao] = null;
            this.b[ao] = null;
        }
        this.c = null;
        var ap = dateNow();
    };
    RearrangedListView.prototype.bu = function (a, b, c) {
        if (b) {
            for (var d = 0; d < this.b.length; d++) {
                this.b[d] = a;
                if (!c) {
                    for (var e = 0; e < this.ah.k.count; e++) {
                        var f = this.ah.k.item(e);
                        this.c[d][e] = f.f();
                        this.c[d][e].beginCalculation(this.n, f.propertyName);
                    }
                }
            }
        }
        if (this.k == 1) {
            var g = false;
            var h = a.e.length;
            for (var i = 0; i < h; i++) {
                if (g) {
                    if (!c) {
                        for (var j = 0; j < this.ah.k.count; j++) {
                            this.b[i].i[i][j] = this.c[i][j].getResults();
                            this.b[i].i[i][j].summaryIndex = j;
                            this.b[i].i[i][j].groupKey = this.b[i].f;
                            this.c[i][j].beginCalculation(this.n, this.ah.k.item(j).propertyName);
                        }
                        this.b[i].h[i] = a.af;
                    }
                    this.b[i] = a;
                    continue;
                }
                var k = false;
                if (this.b[i].f[i] == null) {
                    k = a.f[i] != null;
                }
                else {
                    k = !Base.equalsStatic(this.b[i].f[i], a.f[i]);
                }
                if (k) {
                    if (!c) {
                        for (var l = 0; l < this.ah.k.count; l++) {
                            this.b[i].i[i][l] = this.c[i][l].getResults();
                            this.b[i].i[i][l].summaryIndex = l;
                            this.b[i].i[i][l].groupKey = this.b[i].f;
                            this.c[i][l].beginCalculation(this.n, this.ah.k.item(l).propertyName);
                        }
                        this.b[i].h[i] = a.af;
                    }
                    this.b[i] = a;
                    if (i - 1 > -1) {
                        var m = i - 1;
                        while (m >= 0 && this.b[m].o) {
                            m = m - 1;
                        }
                        a.z = this.b[m].ae;
                    }
                    g = true;
                }
            }
        }
        else {
            var n = false;
            for (var o = 0; o < a.e.length; o++) {
                if (this.b[0].f[o] == null) {
                    if (this.b[0].f[o] != a.f[o]) {
                        n = true;
                        break;
                    }
                    continue;
                }
                if (!Base.equalsStatic(this.b[0].f[o], a.f[o])) {
                    n = true;
                    break;
                }
            }
            if (n) {
                if (!c) {
                    for (var p = 0; p < this.ah.k.count; p++) {
                        this.b[0].i[0][p] = this.c[0][p].getResults();
                        this.b[0].i[0][p].summaryIndex = p;
                        this.b[0].i[0][p].groupKey = this.b[0].f;
                        this.c[0][p].beginCalculation(this.n, this.ah.k.item(p).propertyName);
                    }
                    this.b[0].h[0] = a.af;
                }
                this.b[0] = a;
            }
        }
    };
    RearrangedListView.prototype.bc = function () {
        var a = 0;
        var b = new Dictionary$2(String_$type, Number_$type, 0);
        for (var c = 0; c < this.ah.k.count; c++) {
            if (!this.ah.k.item(c).shouldDisplay) {
                continue;
            }
            var d = this.ah.k.item(c).propertyName;
            if (!b.containsKey(d)) {
                b.addItem(d, 1);
            }
            else {
                b.item(d, b.item(d) + 1);
            }
            if (b.item(d) > a) {
                a = b.item(d);
            }
        }
        return a;
    };
    RearrangedListView.prototype.bf = function (a, b, c, d, e, f) {
        var g = 0;
        if (this.aw) {
            if (this.n.sectionHeaderDisplayMode == 0) {
                if (this.ao(a, 0, c, d, e, f)) {
                    this.aa.y();
                    g++;
                    if (b != null) {
                        var h = this.bo(b);
                        b.a[0] = h;
                        b.c[0] = this.am(h);
                        b.b[0] = (a.count - 1) - b.af;
                    }
                }
            }
            else if (this.n.sectionHeaderDisplayMode == 1) {
                for (var i = 0, j = 0; i < e.length; i++) {
                    var k = a.count + j;
                    if (this.ao(a, i, c, d, [e[i]], f)) {
                        g++;
                        this.aa.y();
                        if (b != null) {
                            var l = this.bp(b, i);
                            b.a[i] = l;
                            b.c[i] = this.am(l);
                            b.b[i] = k - b.af;
                            if (!b.c[i]) {
                                break;
                            }
                        }
                        else if (!this.as) {
                            break;
                        }
                        j++;
                    }
                }
            }
        }
        return g;
    };
    RearrangedListView.prototype.be = function (a, b, c, d, e) {
        var f = 0;
        if (this.aw) {
            var g = this.a6._inner[b];
            if (this.n.sectionHeaderDisplayMode == 0) {
                if (this.ap(b, 0, g, c, d, e)) {
                    this.aa.y();
                    f++;
                    if (a != null) {
                        var h = this.bo(a);
                        a.a[0] = h;
                        a.c[0] = this.am(h);
                        a.b[0] = b - a.af;
                    }
                }
            }
            else if (this.n.sectionHeaderDisplayMode == 1) {
                for (var i = 0, j = 0; i < d.length; i++) {
                    var k = b + j;
                    if (this.ap(k, i, g, c, [d[i]], e)) {
                        f++;
                        this.aa.y();
                        if (a != null) {
                            var l = this.bp(a, i);
                            a.a[i] = l;
                            a.c[i] = this.am(l);
                            a.b[i] = k - a.af;
                            if (!a.c[i]) {
                                break;
                            }
                        }
                        else if (!this.as) {
                            break;
                        }
                        j++;
                    }
                }
            }
        }
        return f;
    };
    RearrangedListView.prototype.ap = function (a, b, c, d, e, f) {
        var g = new DataSourceSpecialRow();
        g.level = b;
        for (var h = 0; h < d.length; h++) {
            g.setValue(d[h], this.n.getItemProperty(c, d[h]));
            if (h < e.length) {
                g.setSectionValue(e[h], this.n.getItemProperty(c, e[h]));
            }
        }
        g.rowType = 1;
        g.summaryResults = new Array(this.ah.k.count);
        if (f && this.a != null) {
            if (this.a[b] != null) {
                if (this.a[b].l(g)) {
                    return false;
                }
                else {
                    this.a[b] = g;
                    for (var i = b + 1; i < this.a.length; i++) {
                        this.a[i] = null;
                    }
                }
            }
            else {
                this.a[b] = g;
            }
        }
        this.a6.insert1(a, g);
        return true;
    };
    RearrangedListView.prototype.ao = function (a, b, c, d, e, f) {
        var g = new DataSourceSpecialRow();
        g.level = b;
        for (var h = 0; h < d.length; h++) {
            g.setValue(d[h], this.n.getItemProperty(c, d[h]));
            if (h < e.length) {
                g.setSectionValue(e[h], this.n.getItemProperty(c, e[h]));
            }
        }
        g.rowType = 1;
        g.summaryResults = new Array(this.ah.k.count);
        if (f && this.a != null) {
            if (this.a[b] != null) {
                if (this.a[b].l(g)) {
                    return false;
                }
                else {
                    this.a[b] = g;
                    for (var i = b + 1; i < this.a.length; i++) {
                        this.a[i] = null;
                    }
                }
            }
            else {
                this.a[b] = g;
            }
        }
        a.add1(g);
        return true;
    };
    RearrangedListView.prototype.bd = function (a, b, c, d) {
        var e = 0;
        if (this.av) {
            if (this.n.sectionHeaderDisplayMode == 0) {
                this.bx(a, 0, b, c, d);
                e++;
            }
            else if (this.n.sectionHeaderDisplayMode == 1) {
                for (var f = d.length - 1; f >= 0; f--) {
                    if (this.aa.t._inner[this.aa.t.count - 1].c[f]) {
                        this.bx(a, f, b, c, [d[f]]);
                        a++;
                        e++;
                    }
                }
            }
        }
        return e;
    };
    RearrangedListView.prototype.bx = function (a, b, c, d, e) {
        var f = new DataSourceSpecialRow();
        f.level = b;
        for (var g = 0; g < d.length; g++) {
            f.setValue(d[g], this.n.getItemProperty(c, d[g]));
            if (g < e.length) {
                f.setSectionValue(e[g], this.n.getItemProperty(c, e[g]));
            }
        }
        f.rowType = 2;
        this.a6.insert1(a, f);
    };
    RearrangedListView.prototype.r = function (a, b, c) {
        var d = this.s(b, 0, c, 5);
        if (this.at) {
            a.ag = a.ag + 1;
        }
        else {
            a.ah = a.ah + 1;
        }
        a.ad += 1;
        this.aa.ac(a, 1, 1);
        return d;
    };
    RearrangedListView.prototype.s = function (a, b, c, d) {
        var e = ((function () {
            var $ret = new DataSourceSpecialRow();
            $ret.level = b;
            $ret.rowType = d;
            $ret.summaryResults = c;
            return $ret;
        })());
        this.a6.insert1(a, e);
        return e;
    };
    RearrangedListView.prototype.bh = function (a, b, c) {
        var d = a.v();
        var e = a.af + a.ah + a.aa;
        if (this.aw) {
            e += d;
        }
        this.bj(e, a.c.length, b);
        a.aa += 1;
        a.ad += 1;
        a.t += 1;
        if (c) {
            this.aa.ac(a, 1, 1);
        }
        return e;
    };
    RearrangedListView.prototype.bi = function (a, b, c) {
        var d = a.v();
        if (this.aw) {
            b += d;
        }
        this.bj(b, a.c.length, c);
        a.aa += 1;
        a.ad += 1;
        a.t += 1;
        this.aa.ac(a, 1, 1);
        return b;
    };
    RearrangedListView.prototype.bj = function (a, b, c) {
        var d = new DataSourceSpecialRow();
        d.rowType = 3;
        d.targetRow = c;
        d.level = b;
        this.a6.insert1(a, d);
        return a + 1;
    };
    RearrangedListView.prototype.bk = function (a, b) {
        var c = -1;
        for (var d = 0; d < this.a5.count; d++) {
            if (this.q.i(b, this.a5._inner[d]) == 0) {
                c = d;
                break;
            }
        }
        return this.bl(a, b, c);
    };
    RearrangedListView.prototype.bl = function (a, b, c) {
        var d = -1;
        if (c != -1) {
            var e = a.af;
            var f = e + a.aa;
            var g = false;
            for (var h = e; h < f; h++) {
                var i = this.aw ? h + 1 : h;
                var j = typeCast(DataSourceSpecialRow.$, this.a6._inner[i]);
                if (j.targetRow != null) {
                    if (this.a2 != null) {
                        var k = this.a2.compare(b, j.targetRow);
                        if (k == 0 || k < 0) {
                            d = this.bi(a, h, b);
                            g = true;
                            break;
                        }
                    }
                    else {
                        var l = -1;
                        for (var m = 0; m < this.a5.count; m++) {
                            if (this.q.i(j.targetRow, this.a5._inner[m]) == 0) {
                                l = m;
                                break;
                            }
                        }
                        if (l > c) {
                            d = this.bi(a, h, b);
                            g = true;
                            break;
                        }
                    }
                }
            }
            if (!g) {
                d = this.bh(a, b, true);
                g = true;
            }
        }
        return d;
    };
    RearrangedListView.prototype.by = function (a) {
        var b = SectionData.r(a);
        this.a9 = b.count;
        for (var c = 0; c < b.count; c++) {
            var d = new Array(b._inner[c].count);
            b._inner[c].values.copyTo(d, 0);
            this.s(this.a6.count, 0, d, 4);
        }
    };
    RearrangedListView.prototype.bz = function () {
        if (this.m == 0 || this.m == 2) {
            for (var a = 0; a < this.aa.t.count; a++) {
                var b = this.aa.t._inner[a];
                if (!b.q) {
                    continue;
                }
                if (this.ay && this.an) {
                    if (this.at) {
                        this.b0(b, false);
                    }
                    else {
                        this.b1(b, false);
                    }
                }
                else {
                    for (var c = 0; c < b.i.length; c++) {
                        var d = this.a6._inner[b.af + b.b[c]];
                        if (d != null) {
                            d.summaryResults = b.i[c];
                        }
                        if (!b.c[c]) {
                            break;
                        }
                    }
                }
            }
        }
    };
    RearrangedListView.prototype.b1 = function (a, b) {
        var c = 0;
        var d = this.bn(a, !this.at);
        for (var e = 0; e < a.i.length; e++) {
            if (a.a[e] == null || !a.c[e]) {
                continue;
            }
            var f = SectionData.r(a.i[e]);
            if (this.aw) {
                c += 1;
            }
            a.g[e] = new SummaryGroup();
            a.g[e].a = new Array(this.ba);
            for (var g = 0; g < f.count; g++) {
                var h = new Array(f._inner[g].count);
                f._inner[g].values.copyTo(h, 0);
                var i = a.af + c + g;
                var j = this.s(i, e + 1, h, 5);
                a.ak(j, g);
                for (var k = e + 1; k < a.b.length; k++) {
                    a.b[k]++;
                }
                a.ad += 1;
                a.ah += 1;
                this.aa.ad(a, 1, 1);
                if (d != null) {
                    var l = this.f();
                    for (var m = 0; m < l.length; m++) {
                        j.setValue(l[m], this.n.getItemProperty(d, l[m]));
                    }
                }
                if (b && this.ac != null) {
                    this.ac.notifyInsertItem(i, j);
                }
            }
            c += f.count;
        }
    };
    RearrangedListView.prototype.bg = function (a, b, c) {
        var d = a.h[b];
        var e = SectionData.r(a.i[b]);
        for (var f = 0; f < e.count; f++) {
            var g = new Array(e._inner[f].count);
            e._inner[f].values.copyTo(g, 0);
            var h = this.aa.e(d - 1);
            var i = this.s(d + f, b + 1, g, 5);
            a.ak(i, f);
            h.ad = h.ad + 1;
            h.ag = h.ag + 1;
            this.aa.ac(h, 1, 1);
            var j = null;
            var k = 0;
            for (var l = d + f - 1; l >= 0; l--) {
                var m = this.n.getRowType(l);
                if (m == 0 || m == 1) {
                    j = this.a6._inner[l];
                    k = m;
                    break;
                }
            }
            if (j != null) {
                var n = this.f();
                if (k == 1) {
                    var o = typeCast(DataSourceSpecialRow.$, j);
                    for (var p = 0; p < n.length; p++) {
                        i.setValue(n[p], o.getValue(n[p]));
                    }
                }
                else {
                    for (var q = 0; q < n.length; q++) {
                        i.setValue(n[q], this.n.getItemProperty(j, n[q]));
                    }
                }
            }
            if (c && this.ac != null) {
                this.ac.notifyInsertItem(d + f, i);
            }
        }
        return e.count;
    };
    RearrangedListView.prototype.b0 = function (a, b) {
        for (var c = 0; c < a.i.length; c++) {
            if (!a.c[c]) {
                break;
            }
            if (a.h[c] == -1) {
                continue;
            }
            a.g[c] = new SummaryGroup();
            a.g[c].a = new Array(this.ba);
            var d = this.bg(a, c, b);
        }
    };
    RearrangedListView.prototype.f = function () {
        var a = new Array(this.af.k.count + this.ag.k.count);
        var b = 0;
        for (var c = 0; c < this.af.k.count; c++) {
            a[b] = this.af.k.item(c).propertyName;
            b++;
        }
        for (var d = 0; d < this.ag.k.count; d++) {
            a[b] = this.ag.k.item(d).propertyName;
            b++;
        }
        return a;
    };
    RearrangedListView.prototype.g = function () {
        var a = new Array(this.af.k.count);
        var b = 0;
        for (var c = 0; c < this.af.k.count; c++) {
            a[b] = this.af.k.item(c).propertyName;
            b++;
        }
        return a;
    };
    RearrangedListView.prototype.bv = function (a, b, c) {
        this.aa.x(a);
    };
    RearrangedListView.prototype.y = function (a, b, c) {
        return this.aa.i(a, this.n, b, c, this.n.sectionHeaderDisplayMode, this.as);
    };
    RearrangedListView.prototype.ce = function (a) {
        this.a5 = new List$1(Array_$type, 0);
        for (var b = 0; b < a.count; b++) {
            this.a5.add(a._inner[b]);
        }
    };
    RearrangedListView.prototype.b4 = function () {
        this.a6.clear();
        if (this.ac != null) {
            this.ac.notifyClearItems();
        }
    };
    RearrangedListView.prototype.notifySetItem = function (a, b, c) {
        if (this.a2 != null) {
            this.cb(b);
            this.bq(c);
        }
        else {
            this.cc(a, b);
            this.br(a, c);
        }
    };
    RearrangedListView.prototype.br = function (a, b) {
        if (!this.az(b)) {
            return;
        }
        var c = this.bb(a);
        this.b5(c, b);
    };
    RearrangedListView.prototype.b5 = function (a, b) {
        this.b2(a, b);
        this.ch(a, b);
        this.cm(a, b);
    };
    RearrangedListView.prototype.b2 = function (a, b) {
        if (this.ai) {
            var c = this.aa.e(a);
            if (c == null) {
                c = this.aa.d();
            }
            var d = c.v();
            var e = true;
            if (a >= c.af && a <= c.af + (d - 1) + c.ah + c.aa) {
                if (!this.aj(c, b, this.a6._inner[c.af])) {
                    c = this.aa.e(c.af - 1);
                    if (c == null) {
                        c = this.aa.c();
                        c.af += 1;
                    }
                    e = false;
                }
                else {
                    a = c.af + d + c.ah + c.aa;
                }
            }
            if (e && !c.p) {
                return;
            }
            c.ad += 1;
            this.aa.ac(c, 1, 0);
            if (this.aj(c, b, this.a6._inner[c.af])) {
                if (!c.p) {
                    c.ad -= 1;
                    this.aa.ac(c, -1, 0);
                    return;
                }
                if (a >= c.ad - c.u - c.ag && a <= c.ad) {
                    a = c.ad - c.u - c.ag;
                }
            }
        }
        this.a6.insert1(a, b);
        if (this.ac != null) {
            this.ac.notifyInsertItem(a, b);
        }
    };
    RearrangedListView.prototype.ch = function (a, b) {
        var c = null;
        if (this.ai) {
            c = this.aa.e(a);
        }
        else {
            c = this.aa.t._inner[0];
        }
        if (c == null) {
            if (this.ai && a == 0) {
                c = this.aa.c();
            }
            else {
                c = this.aa.d();
            }
        }
        if (this.aj(c, b, this.a6._inner[c.af])) {
            this.aa.z(c, 1);
            if (this.ax && c.p) {
                var d = this.bk(c, b);
                if (d != -1 && this.ac != null) {
                    this.ac.notifyInsertItem(d, this.a6._inner[d]);
                }
            }
        }
        else {
            var e = this.aa.e(a - 1);
            if (e != null && this.aj(e, b, this.a6._inner[e.af])) {
                this.aa.z(e, 1);
                if (this.ax && e.p) {
                    var f = this.bk(e, b);
                    if (f != -1 && this.ac != null) {
                        this.ac.notifyInsertItem(f, this.a6._inner[f]);
                    }
                }
            }
            else {
                var g = false;
                if (a <= c.af) {
                    g = true;
                }
                var h = 0;
                var i = null;
                var j = null;
                var k = this.f();
                var l = this.g();
                var m = this.aa.b(this.n, b, l, this.k, this.as);
                m.af = a;
                m.ad = a;
                m.ae = g ? c.ae : c.ae + 1;
                if (this.k == 1) {
                    this.aa.af(m, g);
                    for (var n = 0; n < c.f.length; n++) {
                        if (Base.equalsStatic(c.f[n], m.f[n]) && !c.c[n]) {
                            this.aa.z(c, 1);
                            return;
                        }
                    }
                }
                this.aa.aa(m.ae, m);
                if (this.aw) {
                    var o = this.be(m, a, k, l, false);
                    m.ad += o;
                    h += o;
                    i = this.a6._inner[a];
                    if (!g) {
                        c.ad -= 1;
                    }
                }
                if (m.p) {
                    if (this.ax) {
                        for (var p = 0; p < this.a5.count; p++) {
                            if (this.q.i(b, this.a5._inner[p]) == 0) {
                                this.bh(m, b, false);
                                h++;
                                break;
                            }
                        }
                    }
                }
                else {
                    this.a6.removeAt(a + h);
                    m.ad -= 1;
                    h--;
                }
                if (this.av) {
                    var q = this.bd(a + (h - 1), b, k, l);
                    if (q > 1) {
                        m.ad += q;
                    }
                    h += q;
                    j = this.a6._inner[a + (h - 1)];
                }
                this.aa.ad(m, h, h);
                if (this.k == 1) {
                    this.ca(m);
                }
                else {
                    if (m.a[0] != null && m.h[0] == -1) {
                        m.h[0] = m.ad + 1;
                    }
                }
                if (m.p) {
                    this.aa.z(m, 1);
                }
                if (this.ac != null) {
                    if (i != null) {
                        this.ac.notifyInsertItem(a, i);
                    }
                    if (j != null) {
                        this.ac.notifyInsertItem(a + (h - 1), j);
                    }
                }
            }
        }
    };
    RearrangedListView.prototype.cp = function (a) {
        var b = this.a6.indexOf1(a);
        if (b >= 0) {
            var c = this.a6._inner[b - 1];
            var d = typeCast(DataSourceSpecialRow.$, c);
            var e = this.f();
            for (var f = 0; f < e.length; f++) {
                if (d == null) {
                    a.setValue(e[f], this.n.getItemProperty(c, e[f]));
                }
                else {
                    a.setValue(e[f], d.getValue(e[f]));
                }
            }
        }
    };
    RearrangedListView.prototype.ca = function (a) {
        var b = false;
        var c = false;
        if (a.ae > 0) {
            var d = this.aa.t._inner[a.ae - 1];
            var e = 0;
            for (var f = 0, g = 0; f < a.e.length;) {
                if (!b && Base.equalsStatic(d.f[f], a.f[f])) {
                    this.a6.removeAt(a.af + g);
                    a.ad -= 1;
                    this.aa.ac(a, -1, -1);
                    if (this.an && (this.m == 2 || this.m == 0)) {
                        for (var h = 0; h < this.ba; h++) {
                            var i = typeCast(DataSourceSpecialRow.$, this.a6._inner[a.af - 1]);
                            if (i != null && i.rowType == 5 && i.level == f + 1) {
                                this.a6.removeAt(a.af - 1);
                                d.ad -= 1;
                                this.aa.ac(d, -1, -1);
                                this.a6.insert1(a.ad + 1, i);
                                e++;
                                this.cp(i);
                            }
                        }
                    }
                    a.a[f] = null;
                    a.b[f] = 0;
                    a.c[f] = true;
                    if (d.a[f] != null) {
                        a.z = d.ae;
                    }
                    g--;
                    c = true;
                }
                else {
                    b = true;
                    a.b[f] = g;
                    if (e > 0) {
                        a.ad += e;
                        this.aa.ac(a, e, e);
                        a.h[f] = a.ad - (e - 1);
                    }
                    else {
                        a.h[f] = a.ad + 1;
                    }
                }
                f++;
                g++;
            }
            if (a.z == -1 && c) {
                a.z = d.z;
            }
        }
        if (a.ae < this.aa.t.count - 1) {
            b = false;
            c = false;
            var j = this.aa.t._inner[a.ae + 1];
            for (var k = 0, l = 0; k < a.e.length; k++, l++) {
                if (!b && Base.equalsStatic(j.f[k], a.f[k])) {
                    if (j.a[k] != null) {
                        this.a6.removeAt(j.af + l);
                        j.ad -= 1;
                        this.aa.ac(j, -1, -1);
                        a.al(k, -1);
                        a.i[k] = j.i[k];
                        a.d[k] = j.d[k];
                        if (this.an && (this.m == 2 || this.m == 0)) {
                            var m = j.g[k].a[0];
                            var n = this.a6.indexOf1(m);
                            a.h[k] = n;
                            var o = this.aa.e(n);
                            var p = j.g[k].a.length;
                            for (var q = 0; q < p; q++) {
                                j.g[k].a[q] = null;
                                this.a6.removeAt(n);
                                o.ad -= 1;
                                if (this.at) {
                                    o.ag -= 1;
                                }
                                else {
                                    o.ah -= 1;
                                }
                            }
                            this.aa.ac(o, -p, -p);
                            a.al(k, -p);
                        }
                        j.h[k] = -1;
                        j.d[k] = 0;
                        j.i[k] = null;
                        j.g[k] = null;
                        j.a[k] = null;
                        j.b[k] = 0;
                        j.c[k] = true;
                        c = true;
                    }
                    l--;
                }
                else {
                    b = true;
                    j.b[k] = l;
                    if (a.a[k] != null && a.h[k] == -1) {
                        a.h[k] = a.ad + 1;
                    }
                }
            }
        }
    };
    RearrangedListView.prototype.aj = function (a, b, c) {
        if (this.af.k.count == 0) {
            return true;
        }
        if (this.a3.compare(b, c) == 0) {
            return true;
        }
        return false;
    };
    RearrangedListView.prototype.h = function (a, b) {
        var c = new Array(a.length);
        for (var d = 0; d < c.length; d++) {
            c[d] = this.n.getItemProperty(b, a[d]);
        }
        return c;
    };
    RearrangedListView.prototype.au = function (a, b) {
        if (a == null && b == null) {
            return true;
        }
        if (a == null && b != null) {
            return false;
        }
        if (a != null && b == null) {
            return false;
        }
        return Base.equalsStatic(a, b);
    };
    RearrangedListView.prototype.bb = function (a) {
        if (this.af.k.count == 0 && this.a5.count == 0 && this.t == null && this.o.actualCount == this.a6.count) {
            return a;
        }
        var b = 0;
        var c = 0;
        var d = this.a6.count;
        while (b < a && c < d) {
            while (c < d) {
                var e = this.a6._inner[c];
                var f = false;
                var viewItem_ = e;
                if ((viewItem_.$$isSpecialRow !== undefined)) {
                    f = true;
                }
                if (!f) {
                    break;
                }
                c++;
            }
            var g = this.o.getItemAtIndex(b);
            if (this.au(g, this.a6._inner[c])) {
                c++;
                b++;
            }
            else {
                b++;
            }
        }
        return c;
    };
    RearrangedListView.prototype.cc = function (a, b) {
        var c = this.bb(a);
        if (c < this.a6.count) {
            if (this.au(b, this.a6._inner[c])) {
                this.b6(c, b);
            }
        }
    };
    RearrangedListView.prototype.b6 = function (a, b) {
        this.a6.removeAt(a);
        if (this.ac != null) {
            this.ac.notifyRemoveItem(a, b);
        }
        var c = this.z(a, b);
        this.cn(a, b, c);
    };
    RearrangedListView.prototype.z = function (a, b) {
        var c = null;
        if (this.ai) {
            c = this.aa.e(a);
        }
        else {
            c = this.aa.t._inner[0];
        }
        if (c != null) {
            this.aa.z(c, -1);
            if (c.p) {
                c.ad -= 1;
                this.aa.ac(c, -1, 0);
                if (this.ax) {
                    for (var d = 0; d < c.aa; d++) {
                        var e = 0;
                        if (this.aw) {
                            e = 1;
                        }
                        e += c.ah;
                        var f = this.a6._inner[c.af + d + e];
                        var g = null;
                        var item_ = f;
                        if ((item_ != null && item_.$$isSpecialRow !== undefined)) {
                            g = f;
                        }
                        if (g != null) {
                            if (g.targetRow == b) {
                                this.a6.removeAt(c.af + d + e);
                                c.ad -= 1;
                                c.aa -= 1;
                                c.t -= 1;
                                this.aa.ac(c, -1, -1);
                                if (this.ac != null) {
                                    this.ac.notifyRemoveItem(c.af + d + e, g);
                                }
                            }
                        }
                    }
                }
            }
            if (this.ai && !c.m() && this.ak(c)) {
                var h = 0;
                if (this.aa.p(c)) {
                    var i = c.w();
                    var j = i;
                    var k = this.af.k.count;
                    var l = this.g();
                    var m = this.aa.t._inner[c.ae + 1];
                    for (var n = 0; n < c.f.length; n++) {
                        if (!Base.equalsStatic(c.f[n], m.f[n])) {
                            j = n;
                            break;
                        }
                    }
                    if (this.aw) {
                        if (c.c[j - 1]) {
                            for (var o = j; o < k; o++) {
                                if (c.g[o] != null) {
                                    for (var p = 0; p < c.g[o].a.length; p++) {
                                        var q = c.g[o].a[p];
                                        if (q != null) {
                                            var r = this.a6.indexOf1(q);
                                            if (r >= 0) {
                                                this.a6.removeAt(r);
                                                c.ad -= 1;
                                                h++;
                                                if (this.ac != null) {
                                                    this.ac.notifyRemoveItem(r, q);
                                                }
                                                if (!this.at) {
                                                    for (var s = o + 1; s < c.b.length; s++) {
                                                        c.b[s]--;
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                                var t = c.af + c.b[o];
                                var u = typeCast(DataSourceSpecialRow.$, this.a6._inner[t]);
                                if (u != null && u.rowType == 1) {
                                    c.a[o] = null;
                                    this.a6.removeAt(t);
                                    c.ad -= 1;
                                    h++;
                                    if (this.ac != null) {
                                        this.ac.notifyRemoveItem(t, u);
                                    }
                                    for (var v = o + 1; v < c.b.length; v++) {
                                        c.b[v]--;
                                    }
                                }
                                if (!c.c[o]) {
                                    break;
                                }
                            }
                        }
                    }
                    c.k = m;
                    for (var w = 0; w < j; w++) {
                        if (this.an && !this.at && (this.m == 2 || this.m == 0)) {
                            m.af -= (this.aw ? this.ba + 1 : this.ba);
                        }
                        else if (this.aw) {
                            m.af -= 1;
                        }
                    }
                    m.d[i] = c.d[i];
                    m.a[i] = c.a[i];
                    m.c[i] = c.c[i];
                    m.z = c.z;
                    for (var x = i; x < k; x++) {
                        m.g[x] = c.g[x];
                    }
                    m.b[i] = c.b[i];
                    for (var y = i, z = c.b[i]; y < k; y++) {
                        m.b[y] = z;
                        if (this.aw) {
                            z++;
                        }
                        if (!this.at) {
                            if (m.g[y] != null) {
                                z += m.g[y].a.length;
                            }
                        }
                        var aa = m.af - h + m.b[y];
                        var ab = typeCast(DataSourceSpecialRow.$, this.a6._inner[aa]);
                        if (ab != null && ab.rowType == 1) {
                            for (var ac = 0; ac < m.f.length; ac++) {
                                ab.setValue(m.e[ac], m.f[ac]);
                            }
                        }
                    }
                }
                else {
                    if (this.aw) {
                        for (var ad = c.af; ad <= c.ad; ad++) {
                            var ae = typeCast(DataSourceSpecialRow.$, this.a6._inner[ad]);
                            if (ae != null && ae.rowType == 1) {
                                this.a6.removeAt(ad);
                                c.ad -= 1;
                                h++;
                                ad--;
                                if (this.ac != null) {
                                    this.ac.notifyRemoveItem(ad, ae);
                                }
                            }
                        }
                    }
                    for (var af = 0; af < c.g.length; af++) {
                        if (c.g[af] == null) {
                            continue;
                        }
                        for (var ag = 0; ag < c.g[af].a.length; ag++) {
                            var ah = c.g[af].a[ag];
                            if (ah != null) {
                                var ai = this.a6.indexOf1(ah);
                                this.a6.removeAt(ai);
                                c.ad -= 1;
                                h++;
                                if (this.ac != null) {
                                    this.ac.notifyRemoveItem(ai, ah);
                                }
                            }
                        }
                    }
                    if (c.z >= 0) {
                        var aj = c.ad - c.af + 1;
                        if (aj > 0) {
                            var ak = this.a6._inner[c.af - 1];
                            var al = typeCast(DataSourceSpecialRow.$, ak);
                            var am = this.f();
                            for (var an = 0; an < aj; an++) {
                                var ao = c.af + an;
                                var ap = typeCast(DataSourceSpecialRow.$, this.a6._inner[ao]);
                                if (ap != null) {
                                    for (var aq = 0; aq < am.length; aq++) {
                                        if (al == null) {
                                            ap.setValue(am[aq], this.n.getItemProperty(ak, am[aq]));
                                        }
                                        else {
                                            ap.setValue(am[aq], al.getValue(am[aq]));
                                        }
                                    }
                                }
                            }
                            var ar = this.aa.t._inner[c.ae - 1];
                            ar.ad += aj;
                        }
                    }
                }
                if (this.av) {
                    for (var as = 0; as < c.a.length; as++) {
                        if (!c.c[as]) {
                            break;
                        }
                        if (c.a[as] != null) {
                            var at = this.a6._inner[c.ad];
                            this.a6.removeAt(c.ad);
                            if (this.ac != null) {
                                this.ac.notifyRemoveItem(c.ad, at);
                            }
                            h++;
                            c.ad -= 1;
                        }
                    }
                }
                this.aa.ae(c.ae);
                this.aa.ab(c, -h, -h);
                c.o = true;
            }
        }
        return c;
    };
    RearrangedListView.prototype.ak = function (a) {
        for (var b = a.af; b <= a.ad; b++) {
            if (this.n.getRowType(b) == 0) {
                return false;
            }
        }
        return true;
    };
    RearrangedListView.prototype.bq = function (a) {
        if (this.az(a)) {
            var b = this.a6.g(a, this.a2);
            if (b < 0) {
                b = ~b;
            }
            this.b5(b, a);
        }
    };
    RearrangedListView.prototype.b8 = function (a) {
        this.a5.add(a);
        var b = this.n.indexOfKey(a);
        var c = this.n.getItemAtIndex(b);
        if (b == -1) {
            return;
        }
        if (this.ax) {
            var d = this.aa.e(b);
            var e = this.bk(d, this.a6._inner[b]);
            if (this.ac != null) {
                this.ac.notifyInsertItem(e, this.a6._inner[e]);
            }
        }
    };
    RearrangedListView.prototype.cf = function (a) {
        if (this.a5.remove(a)) {
            var b = this.n.indexOfKey(a);
            var c = this.n.getItemAtIndex(b);
            if (b == -1) {
                return;
            }
            var d = this.aa.e(b);
            var e = d.af;
            if (this.aw) {
                e += d.v();
            }
            if (this.ax) {
                var f = this.v(e, this.a6._inner[b], d.aa, this.f());
                if (f != null) {
                    d.ad -= 1;
                    d.aa -= 1;
                    d.t -= 1;
                    this.aa.ac(d, -1, -1);
                    if (this.ac != null) {
                        this.ac.notifyRemoveItem(f.a, f.item);
                    }
                }
            }
        }
    };
    RearrangedListView.prototype.v = function (a, b, c, d) {
        var _loop_1 = function (e) {
            if (typeCast(DataSourceSpecialRow.$, this_1.a6._inner[e]) !== null) {
                var f_1 = this_1.a6._inner[e];
                if (f_1.rowType == 3) {
                    var g = f_1.targetRow;
                    if (g == b) {
                        this_1.a6.removeAt(e);
                        return { value: ((function () {
                                var $ret = new RemovedShiftedRowInfo();
                                $ret.item = f_1;
                                $ret.a = e;
                                return $ret;
                            })()) };
                    }
                }
            }
        };
        var this_1 = this;
        for (var e = a + (c - 1); e >= a; e--) {
            var state_1 = _loop_1(e);
            if (typeof state_1 === "object")
                return state_1.value;
        }
        return null;
    };
    RearrangedListView.prototype.cb = function (a) {
        if (!this.az(a)) {
            return;
        }
        var b = this.a6.g(a, this.a2);
        if (b < 0) {
            b = ~b;
        }
        if (b >= 0) {
            var c = this.a6.count;
            var d = 0;
            var e = false;
            while (d <= c) {
                var f = b - d;
                var g = b + d;
                if (f < 0 && g > c - 1) {
                    break;
                }
                if (f >= 0 && f < c) {
                    if (this.a6._inner[f] != null && Base.equalsStatic(this.a6._inner[f], a)) {
                        b = f;
                        e = true;
                        break;
                    }
                }
                if (g < c) {
                    if (this.a6._inner[g] != null && Base.equalsStatic(this.a6._inner[g], a)) {
                        b = g;
                        e = true;
                        break;
                    }
                }
                d++;
            }
            if (b < 0 || b > c - 1) {
                return;
            }
            if (e) {
                var h = this.a6._inner[b];
                if (h == null && a == null) {
                    this.a6.removeAt(b);
                    this.ab.notifyRemoveItem(b, a);
                    var i = this.z(b, a);
                    this.cl(b, a, false, i);
                    return;
                }
                if (h == null) {
                    return;
                }
                if (Base.equalsStatic(h, a)) {
                    this.a6.removeAt(b);
                    this.ab.notifyRemoveItem(b, a);
                    var j = this.z(b, a);
                    this.cl(b, a, false, j);
                }
            }
            else {
                if (b - 1 >= 0) {
                    var k = this.aa.e(b - 1);
                    if (k != null && !k.p) {
                        b = b - 1;
                    }
                }
                this.z(b, a);
            }
        }
        else {
            var l = this.a6.count;
            for (var m = 0; m < l; m++) {
                var n = this.a6._inner[m];
                if (n == null && a == null) {
                    this.a6.removeAt(m);
                    break;
                }
                if (n == null) {
                    continue;
                }
                if (Base.equalsStatic(n, a)) {
                    this.a6.removeAt(m);
                    this.ab.notifyRemoveItem(m, a);
                    var o = this.z(m, a);
                    this.cn(m, a, o);
                    break;
                }
            }
        }
    };
    RearrangedListView.prototype.az = function (a) {
        if (this.t == null) {
            return true;
        }
        return this.t(a);
    };
    RearrangedListView.prototype.notifyClearItems = function () {
    };
    RearrangedListView.prototype.notifyInsertItem = function (a, b) {
        if (this.a2 != null) {
            this.bq(b);
        }
        else {
            this.br(a, b);
        }
    };
    RearrangedListView.prototype.notifyRemoveItem = function (a, b) {
        if (this.a2 != null) {
            this.cb(b);
        }
        else {
            this.cc(a, b);
        }
    };
    RearrangedListView.prototype.cd = function (a, b) {
        var c = this.aa.e(a);
        if (c == null) {
            c = this.aa.d();
        }
        for (var d = 0, e = 0; d < c.a.length; d++) {
            if (c.a[d] != null) {
                if (a == c.af + c.b[d]) {
                    var f = c.a[d];
                    if (this.a0.containsKey(f)) {
                        if (this.a0.item(f) != b) {
                            var g = this.a0.item(f);
                            if (b == this.as) {
                                this.a0.removeItem(f);
                            }
                            else {
                                this.a0.item(f, b);
                            }
                            this.n.queueAutoRefresh();
                            if (this.ae != null) {
                                this.ae.notifyRowExpansionChanged(a, g, b);
                            }
                        }
                    }
                    else if (b != this.as) {
                        this.a0.addItem(f, b);
                        this.n.queueAutoRefresh();
                        if (this.ae != null) {
                            this.ae.notifyRowExpansionChanged(a, this.as, b);
                        }
                    }
                }
                e++;
            }
        }
    };
    RearrangedListView.prototype.al = function (a) {
        var b = this.aa.e(a);
        if (b == null) {
            b = this.aa.d();
        }
        for (var c = 0, d = 0; c < b.a.length; c++) {
            if (b.a[c] != null) {
                if (a == b.af + b.b[c]) {
                    return b.c[c];
                }
                d++;
            }
        }
        return true;
    };
    RearrangedListView.prototype.am = function (a) {
        if (!this.ar) {
            return true;
        }
        return this.a0.containsKey(a) ? this.a0.item(a) : this.as;
    };
    RearrangedListView.prototype.bo = function (a) {
        var b = "";
        for (var c = 0; c < a.e.length; c++) {
            var d = a.e[c];
            var e = a.f[c] != null ? a.f[c].toString() : "";
            b += d + ":" + e;
        }
        return b;
    };
    RearrangedListView.prototype.bp = function (a, b) {
        if (this.n.sectionHeaderDisplayMode == 0) {
            return this.bo(a);
        }
        var c = "";
        for (var d = 0; d < b + 1; d++) {
            var e = a.e[d];
            var f = a.f[d] != null ? a.f[d].toString() : "";
            c += e + ":" + f + ":";
        }
        return c;
    };
    RearrangedListView.prototype.bt = function (a) {
        if (this.m == 3 || (a != null && this.m == 1) || (a == null && this.m == 2)) {
            return;
        }
        if (a == null && (this.e == null || this.e.length != this.ah.k.count)) {
            this.e = new Array(this.ah.k.count);
        }
        if (a != null) {
            var b = a.v();
            for (var c = b - 1; c >= 0; c--) {
                for (var d = 0; d < this.ah.k.count; d++) {
                    var e = this.ah.k.item(d);
                    this.c[c][d].beginCalculation(this.n, e.propertyName);
                }
            }
        }
        else {
            if (this.d == null || this.d.length != this.ah.k.count) {
                this.d = new Array(this.ah.k.count);
            }
            for (var f = 0; f < this.ah.k.count; f++) {
                var g = this.ah.k.item(f);
                if (this.d[f] == null || this.d[f].operand != g.operand) {
                    this.d[f] = g.f();
                }
                this.d[f].beginCalculation(this.n, g.propertyName);
            }
        }
    };
    RearrangedListView.prototype.bw = function (a) {
        if (this.m == 3 || (a != null && this.m == 1) || (a == null && this.m == 2)) {
            return;
        }
        if (a != null) {
            var b = a.v();
            for (var c = b - 1; c >= 0; c--) {
                var d = this.a6._inner[a.af + c];
                for (var e = 0; e < this.ah.k.count; e++) {
                    var f = this.ah.k.item(e);
                    var g = this.c[c][e].getResults();
                    g.summaryIndex = e;
                    g.groupKey = a.f;
                    if (d != null) {
                        d.summaryResults[c] = g;
                    }
                }
            }
        }
        else {
            for (var h = 0; h < this.ah.k.count; h++) {
                var i = this.d[h].getResults();
                i.summaryIndex = h;
                this.e[h] = i;
            }
        }
    };
    RearrangedListView.prototype.bs = function (a, b) {
        if (this.m == 3 || (a != null && this.m == 1) || (a == null && this.m == 2)) {
            return;
        }
        if (b != null) {
            if (a != null) {
                for (var c = 0; c < this.c.length; c++) {
                    if (this.c[c] != null) {
                        for (var d = 0; d < this.c[c].length; d++) {
                            if (this.c[c][d] != null) {
                                this.c[c][d].aggregate(b);
                            }
                        }
                    }
                }
            }
            else {
                for (var e = 0; e < this.ah.k.count; e++) {
                    this.d[e].aggregate(b);
                }
            }
        }
    };
    RearrangedListView.prototype.cj = function (a, b, c, d, e) {
        if (this.k == 0) {
            d = 0;
        }
        var f = a.i[d][e];
        var g = b.itemAdded(c, f, a.d[d]);
        a.an(d, e, g);
        if (this.aw) {
            if (a.a[d] != null) {
                var h = typeCast(DataSourceSpecialRow.$, this.a6._inner[a.af + a.b[d]]);
                if (h != null) {
                    h.summaryResults[e] = a.i[d][e];
                }
            }
        }
    };
    RearrangedListView.prototype.ci = function (a, b, c, d, e, f) {
        if (this.k == 0) {
            d = 0;
        }
        var g = a.i[d][e];
        var h = a.d[d];
        var i = f ? b.itemAdded(c, g, h) : b.itemRemoved(c, g, h);
        a.an(d, e, i);
        if (this.aw) {
            if (!a.o && a.a[d] != null) {
                var j = typeCast(DataSourceSpecialRow.$, this.a6._inner[a.af + a.b[d]]);
                if (j != null) {
                    j.summaryResults[e] = a.i[d][e];
                }
            }
        }
    };
    RearrangedListView.prototype.cl = function (a, b, c, d) {
        if (this.ah.k.count == 0) {
            return;
        }
        if (d == null) {
            d = this.x(a);
            if (d == null) {
                d = this.aa.d();
            }
        }
        var e = false;
        var f = false;
        var g = false;
        for (var h = 0; h < this.ah.k.count; h++) {
            var i = this.ah.k.item(h);
            if (this.m == 0 || this.m == 1) {
                var j = null;
                if (c) {
                    j = this.d[h].itemAdded(b, this.e[h], this.o.actualCount);
                }
                else {
                    j = this.d[h].itemRemoved(b, this.e[h], this.o.actualCount);
                }
                j.summaryIndex = h;
                this.e[h] = j;
                if (this.d[h].isSummaryDirty) {
                    e = true;
                }
                else {
                    f = true;
                }
            }
            if (this.m == 0 || this.m == 2) {
                if (d != null && d.i != null) {
                    var k = i.f();
                    k.dataSource = this.n;
                    k.propertyName = i.propertyName;
                    k.propertyType = this.n.resolveSchemaPropertyType(i.propertyName);
                    var l = this.g();
                    var m = d.w();
                    for (var n = m; n < l.length; n++) {
                        this.ci(d, k, b, n, h, c);
                        if (k.isSummaryDirty) {
                            e = true;
                        }
                        g = true;
                    }
                    var o = d.z;
                    while (o != -1) {
                        var p = this.aa.t._inner[o];
                        var q = p.w();
                        for (var r = q; r < m; r++) {
                            this.ci(p, k, b, r, h, c);
                            if (k.isSummaryDirty) {
                                e = true;
                            }
                        }
                        o = p.z;
                        m = m - 1;
                    }
                }
            }
        }
        if (e) {
            this.n.queueAutoRefresh();
        }
        else {
            if (f) {
                var s = SectionData.r(this.e);
                this.cg(s);
            }
            if (g) {
                this.ck(d, null);
                var t = d.z;
                while (t != -1) {
                    var u = this.aa.t._inner[t];
                    this.ck(u, null);
                    t = u.z;
                }
            }
        }
        if (f) {
            this.b7();
        }
    };
    RearrangedListView.prototype.co = function (a, b, c, d, e) {
        var f = a;
        var g = this.ah.k.item(a);
        if (g == null) {
            return;
        }
        if (e == null) {
            e = this.x(b);
            if (e == null) {
                e = this.aa.d();
            }
        }
        var h = false;
        var i = false;
        var j = false;
        if (this.m == 0 || this.m == 1) {
            var k = null;
            if (d) {
                k = this.d[f].itemAdded(c, this.e[f], this.o.actualCount);
            }
            else {
                k = this.d[f].itemRemoved(c, this.e[f], this.o.actualCount);
            }
            k.summaryIndex = f;
            this.e[f] = k;
            if (this.d[f].isSummaryDirty) {
                h = true;
            }
            else {
                i = true;
            }
        }
        if (this.m == 0 || this.m == 2) {
            if (e != null && e.i != null) {
                var l = g.f();
                l.dataSource = this.n;
                l.propertyName = g.propertyName;
                l.propertyType = this.n.resolveSchemaPropertyType(g.propertyName);
                var m = this.g();
                var n = e.w();
                for (var o = n; o < m.length; o++) {
                    this.ci(e, l, c, o, f, d);
                    if (l.isSummaryDirty) {
                        h = true;
                    }
                    j = true;
                }
                var p = e.z;
                while (p != -1) {
                    var q = this.aa.t._inner[p];
                    var r = q.w();
                    for (var s = r; s < n; s++) {
                        this.ci(q, l, c, s, f, d);
                        if (l.isSummaryDirty) {
                            h = true;
                        }
                    }
                    p = q.z;
                    n = n - 1;
                }
            }
        }
        if (h) {
            this.n.queueAutoRefresh();
        }
        else {
            if (i) {
                var t = SectionData.r(this.e);
                this.cg(t);
            }
            if (j) {
                this.ck(e, null);
                var u = e.z;
                while (u != -1) {
                    var v = this.aa.t._inner[u];
                    this.ck(v, null);
                    u = v.z;
                }
            }
        }
        if (i) {
            this.b7();
        }
    };
    RearrangedListView.prototype.cm = function (a, b) {
        if (this.ah.k.count == 0) {
            return;
        }
        var c = this.x(a);
        if (c == null) {
            c = this.aa.d();
        }
        var d = null;
        if (c != null) {
            d = typeCast(DataSourceSpecialRow.$, this.a6._inner[c.af]);
        }
        var e = false;
        var f = false;
        var g = false;
        for (var h = 0; h < this.ah.k.count; h++) {
            var i = this.ah.k.item(h);
            if (this.m == 0 || this.m == 1) {
                var j = this.d[h].itemAdded(b, this.e[h], this.o.actualCount);
                j.summaryIndex = h;
                this.e[h] = j;
                if (this.d[h].isSummaryDirty) {
                    e = true;
                }
                else {
                    f = true;
                }
            }
            if (this.m == 0 || this.m == 2) {
                if (c != null && c.i != null) {
                    var k = i.f();
                    k.dataSource = this.n;
                    k.propertyName = i.propertyName;
                    k.propertyType = this.n.resolveSchemaPropertyType(i.propertyName);
                    var l = this.g();
                    var m = c.w();
                    if (this.k == 1) {
                        for (var n = m; n < l.length; n++) {
                            this.cj(c, k, b, n, h);
                            if (k.isSummaryDirty) {
                                e = true;
                            }
                            g = true;
                        }
                    }
                    else {
                        this.cj(c, k, b, m, h);
                        if (k.isSummaryDirty) {
                            e = true;
                        }
                        g = true;
                    }
                    var o = c.z;
                    while (o != -1) {
                        var p = this.aa.t._inner[o];
                        var q = p.w();
                        for (var r = q; r < m; r++) {
                            this.cj(p, k, b, r, h);
                            if (k.isSummaryDirty) {
                                e = true;
                            }
                        }
                        o = p.z;
                        m = m - 1;
                    }
                }
            }
        }
        if (e) {
            this.n.queueAutoRefresh();
        }
        else {
            if (f) {
                var s = SectionData.r(this.e);
                this.cg(s);
            }
            if (g) {
                this.ck(c, null);
                var t = c.z;
                while (t != -1) {
                    var u = this.aa.t._inner[t];
                    this.ck(u, null);
                    t = u.z;
                }
            }
        }
        if (f) {
            this.b7();
        }
    };
    RearrangedListView.prototype.cn = function (a, b, c) {
        if (this.ah.k.count == 0) {
            return;
        }
        var d = null;
        if (c != null) {
            d = typeCast(DataSourceSpecialRow.$, this.a6._inner[c.af]);
        }
        var e = false;
        var f = false;
        var g = false;
        for (var h = 0; h < this.ah.k.count; h++) {
            var i = this.ah.k.item(h);
            if (this.m == 0 || this.m == 1) {
                var j = this.d[h].itemRemoved(b, this.e[h], this.n.actualCount);
                j.summaryIndex = h;
                this.e[h] = j;
                if (this.d[h].isSummaryDirty) {
                    e = true;
                }
                else {
                    f = true;
                }
            }
            if (this.m == 0 || this.m == 2) {
                if (c != null && c.i != null) {
                    var k = i.f();
                    k.dataSource = this.n;
                    k.propertyName = i.propertyName;
                    k.propertyType = this.n.resolveSchemaPropertyType(i.propertyName);
                    var l = this.g();
                    var m = c.w();
                    if (this.k == 1) {
                        for (var n = m; n < l.length; n++) {
                            this.ci(c, k, b, n, h, false);
                            if (k.isSummaryDirty) {
                                e = true;
                            }
                            g = true;
                        }
                    }
                    else {
                        this.ci(c, k, b, m, h, false);
                        if (k.isSummaryDirty) {
                            e = true;
                        }
                        g = true;
                    }
                    var o = c.z;
                    while (o != -1) {
                        var p = this.aa.t._inner[o];
                        var q = p.w();
                        for (var r = q; r < m; r++) {
                            this.ci(p, k, b, r, h, false);
                            if (k.isSummaryDirty) {
                                e = true;
                            }
                        }
                        o = p.z;
                        m = m - 1;
                    }
                }
            }
        }
        if (e) {
            this.n.queueAutoRefresh();
        }
        else {
            if (f) {
                var s = SectionData.r(this.e);
                this.cg(s);
            }
            if (g) {
                this.ck(c, null);
                var t = c.z;
                while (t != -1) {
                    var u = this.aa.t._inner[t];
                    this.ck(u, null);
                    t = u.z;
                }
            }
        }
        if (f) {
            this.b7();
        }
    };
    RearrangedListView.prototype.b7 = function () {
        if (this.rootSummariesChanged != null) {
            this.rootSummariesChanged(this, new DataSourceRootSummariesChangedEventArgs());
        }
    };
    RearrangedListView.prototype.ck = function (a, b) {
        var _this = this;
        if (this.m == 2 || this.m == 0) {
            if (this.an) {
                if (!a.o && !a.n()) {
                    if (this.at) {
                        this.b0(a, true);
                    }
                    else {
                        this.b1(a, true);
                    }
                }
                else {
                    a.am(function (c) {
                        if (_this.ac != null) {
                            var d = _this.a6.indexOf1(c);
                            _this.ac.notifySetItem(d, c, c);
                        }
                    });
                }
            }
            else {
                if (!a.o) {
                    for (var c = 0; c < a.i.length; c++) {
                        var d = a.af + a.b[c];
                        var e = this.a6._inner[d];
                        if (e != null) {
                            e.summaryResults = a.i[c];
                            if (this.ac != null) {
                                this.ac.notifySetItem(d, e, e);
                            }
                        }
                    }
                }
            }
        }
    };
    RearrangedListView.prototype.cg = function (a) {
        for (var b = 0; b < a.count; b++) {
            var c = new Array(a._inner[b].count);
            a._inner[b].values.copyTo(c, 0);
            var d = this.a6.count - a.count + b;
            var e = typeCast(DataSourceSpecialRow.$, this.a6._inner[d]);
            e.summaryResults = c;
            if (this.ac != null) {
                this.ac.notifySetItem(d, e, e);
            }
        }
    };
    RearrangedListView.$t = markType(RearrangedListView, 'RearrangedListView', Base.$, [ISupportsDataChangeNotifications_$type]);
    return RearrangedListView;
}(Base));
export { RearrangedListView };
