// igniteui-react-core/esm5/lib/ResponsiveStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ResponsiveStateDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ResponsiveStateDescription, _super);
    function ResponsiveStateDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.k = 0;
        _this.j = 0;
        _this.h = false;
        _this.p = null;
        _this.o = null;
        _this.q = null;
        return _this;
    }
    ResponsiveStateDescription.prototype.get_type = function () {
        return "ResponsiveState";
    };
    Object.defineProperty(ResponsiveStateDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ResponsiveStateDescription.prototype, "name", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ResponsiveStateDescription.prototype, "minimumWidth", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("MinimumWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ResponsiveStateDescription.prototype, "maximumWidth", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("MaximumWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ResponsiveStateDescription.prototype, "isManualState", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("IsManualState");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ResponsiveStateDescription.prototype, "stateEnteringRef", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("StateEnteringRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ResponsiveStateDescription.prototype, "stateEnteredRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("StateEnteredRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ResponsiveStateDescription.prototype, "stateExitedRef", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("StateExitedRef");
        },
        enumerable: false,
        configurable: true
    });
    ResponsiveStateDescription.$t = markType(ResponsiveStateDescription, 'ResponsiveStateDescription', Description.$);
    return ResponsiveStateDescription;
}(Description));
export { ResponsiveStateDescription };
