// igniteui-react-core/esm5/lib/RingSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var RingSeriesBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RingSeriesBaseDescription, _super);
    function RingSeriesBaseDescription() {
        var _this = _super.call(this) || this;
        _this.bk = null;
        _this.au = null;
        _this.v = 0;
        _this.aw = null;
        _this.av = null;
        _this.at = null;
        _this.bl = null;
        _this.a1 = null;
        _this.a9 = null;
        _this.a3 = null;
        _this.a7 = null;
        _this.a4 = null;
        _this.a5 = null;
        _this.z = 0;
        _this.y = 0;
        _this.a6 = null;
        _this.x = 0;
        _this.ac = 0;
        _this.bf = null;
        _this.be = null;
        _this.bb = null;
        _this.ax = null;
        _this.ay = null;
        _this.az = null;
        _this.i = null;
        _this.bg = null;
        _this.l = null;
        _this.a8 = null;
        _this.j = null;
        _this.ba = null;
        _this.k = null;
        _this.w = 0;
        _this.ag = 0;
        _this.bc = null;
        _this.bd = null;
        _this.ab = 0;
        _this.aa = 0;
        _this.bi = null;
        _this.bj = null;
        _this.af = 0;
        _this.ae = 0;
        _this.h = null;
        _this.m = null;
        _this.a2 = null;
        _this.a0 = null;
        _this.t = false;
        _this.ad = 0;
        _this.bh = null;
        return _this;
    }
    Object.defineProperty(RingSeriesBaseDescription.prototype, "tooltipTemplateRef", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("TooltipTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "fontFamily", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("FontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "fontSize", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("FontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "fontWeight", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("FontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "fontStyle", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("FontStyle");
        },
        enumerable: false,
        configurable: true
    });
    RingSeriesBaseDescription.prototype.get_type = function () {
        return "RingSeriesBase";
    };
    Object.defineProperty(RingSeriesBaseDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "labelMemberPath", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("LabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "legendLabelMemberPath", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("LegendLabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "labelsPosition", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("LabelsPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "leaderLineVisibility", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("LeaderLineVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "leaderLineFill", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("LeaderLineFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "leaderLineStroke", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("LeaderLineStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "leaderLineStrokeThickness", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("LeaderLineStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "leaderLineOpacity", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("LeaderLineOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "leaderLineType", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("LeaderLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "leaderLineMargin", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("LeaderLineMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "othersCategoryThreshold", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("OthersCategoryThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "othersCategoryType", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("OthersCategoryType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "othersCategoryText", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.g("OthersCategoryText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "legendRef", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("LegendRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "formatLabelRef", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("FormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "formatLegendLabelRef", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("FormatLegendLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "labelFormat", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("LabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "labelFormatSpecifiers", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("LabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "othersLabelFormat", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.g("OthersLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "othersLabelFormatSpecifiers", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("OthersLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "legendLabelFormat", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("LegendLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "legendLabelFormatSpecifiers", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("LegendLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "legendOthersLabelFormat", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("LegendOthersLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "legendOthersLabelFormatSpecifiers", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("LegendOthersLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "labelExtent", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("LabelExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "startAngle", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("StartAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "othersCategoryFill", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("OthersCategoryFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "othersCategoryStroke", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("OthersCategoryStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "othersCategoryStrokeThickness", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("OthersCategoryStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "othersCategoryOpacity", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("OthersCategoryOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "selectedSliceFill", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("SelectedSliceFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "selectedSliceStroke", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("SelectedSliceStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "selectedSliceStrokeThickness", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("SelectedSliceStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "selectedSliceOpacity", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("SelectedSliceOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "brushes", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Brushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "outlines", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Outlines");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "labelOuterColor", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("LabelOuterColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "labelInnerColor", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("LabelInnerColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "isSurfaceInteractionDisabled", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("IsSurfaceInteractionDisabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "radiusFactor", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("RadiusFactor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingSeriesBaseDescription.prototype, "propertyUpdatedRef", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.g("PropertyUpdatedRef");
        },
        enumerable: false,
        configurable: true
    });
    RingSeriesBaseDescription.$t = markType(RingSeriesBaseDescription, 'RingSeriesBaseDescription', Description.$);
    return RingSeriesBaseDescription;
}(Description));
export { RingSeriesBaseDescription };
