// igniteui-react-core/esm5/lib/RingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var RingDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RingDescription, _super);
    function RingDescription() {
        var _this = _super.call(this) || this;
        _this.r = 0;
        _this.n = 0;
        _this.l = null;
        _this.h = null;
        _this.j = null;
        _this.o = 0;
        return _this;
    }
    RingDescription.prototype.get_type = function () {
        return "Ring";
    };
    Object.defineProperty(RingDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingDescription.prototype, "index", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Index");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingDescription.prototype, "innerExtend", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("InnerExtend");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingDescription.prototype, "controlSize", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ControlSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingDescription.prototype, "center", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Center");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingDescription.prototype, "ringSeries", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("RingSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RingDescription.prototype, "ringBreadth", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("RingBreadth");
        },
        enumerable: false,
        configurable: true
    });
    RingDescription.$t = markType(RingDescription, 'RingDescription', Description.$);
    return RingDescription;
}(Description));
export { RingDescription };
