// igniteui-react-core/esm5/lib/RippleDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var RippleDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RippleDescription, _super);
    function RippleDescription() {
        var _this = _super.call(this) || this;
        _this.r = null;
        _this.t = null;
        _this.s = null;
        _this.p = null;
        _this.q = null;
        _this.n = 0;
        _this.i = false;
        _this.h = false;
        _this.j = false;
        return _this;
    }
    RippleDescription.prototype.get_type = function () {
        return "Ripple";
    };
    Object.defineProperty(RippleDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RippleDescription.prototype, "eventSourceRef", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("EventSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RippleDescription.prototype, "rippleColor", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("RippleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RippleDescription.prototype, "hoverColor", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("HoverColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RippleDescription.prototype, "actualHoverColor", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("ActualHoverColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RippleDescription.prototype, "actualRippleColor", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("ActualRippleColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RippleDescription.prototype, "rippleDuration", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("RippleDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RippleDescription.prototype, "isDisabled", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("IsDisabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RippleDescription.prototype, "isCentered", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("IsCentered");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RippleDescription.prototype, "isHoverEnabled", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("IsHoverEnabled");
        },
        enumerable: false,
        configurable: true
    });
    RippleDescription.$t = markType(RippleDescription, 'RippleDescription', Description.$);
    return RippleDescription;
}(Description));
export { RippleDescription };
