// igniteui-react-core/esm5/lib/ScatterBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ScatterBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterBaseDescription, _super);
    function ScatterBaseDescription() {
        var _this = _super.call(this) || this;
        _this.gh = null;
        _this.gl = null;
        _this.gk = null;
        _this.go = null;
        _this.gb = null;
        _this.gc = null;
        _this.gi = null;
        _this.gm = null;
        _this.gj = null;
        _this.gn = null;
        _this.gg = null;
        _this.gf = null;
        _this.f8 = null;
        _this.fx = 0;
        _this.fr = null;
        _this.f1 = 0;
        _this.ge = null;
        _this.f2 = 0;
        _this.f0 = 0;
        _this.gd = null;
        _this.fz = 0;
        _this.f7 = null;
        _this.fu = false;
        _this.ft = false;
        _this.ga = null;
        _this.f9 = null;
        return _this;
    }
    ScatterBaseDescription.prototype.get_type = function () {
        return "ScatterBase";
    };
    Object.defineProperty(ScatterBaseDescription.prototype, "xAxisRef", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("XAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "yAxisRef", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("YAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "xMemberPath", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("XMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "yMemberPath", {
        get: function () {
            return this.go;
        },
        set: function (a) {
            this.go = a;
            this.g("YMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "highlightedXMemberPath", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("HighlightedXMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "highlightedYMemberPath", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("HighlightedYMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "xMemberAsLegendLabel", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("XMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "yMemberAsLegendLabel", {
        get: function () {
            return this.gm;
        },
        set: function (a) {
            this.gm = a;
            this.g("YMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "xMemberAsLegendUnit", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("XMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "yMemberAsLegendUnit", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("YMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "trendLineType", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.f8;
        },
        set: function (a) {
            this.f8 = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "markerCollisionAvoidance", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("MarkerCollisionAvoidance");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "trendLineZIndex", {
        get: function () {
            return this.f2;
        },
        set: function (a) {
            this.f2 = a;
            this.g("TrendLineZIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "maximumMarkers", {
        get: function () {
            return this.f0;
        },
        set: function (a) {
            this.f0 = a;
            this.g("MaximumMarkers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "itemSearchMode", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("ItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "itemSearchThreshold", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("ItemSearchThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "actualItemSearchMode", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("ActualItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "isCustomScatterStyleAllowed", {
        get: function () {
            return this.fu;
        },
        set: function (a) {
            this.fu = a;
            this.g("IsCustomScatterStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "isCustomScatterMarkerStyleAllowed", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("IsCustomScatterMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "assigningScatterStyleRef", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("AssigningScatterStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBaseDescription.prototype, "assigningScatterMarkerStyleRef", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("AssigningScatterMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    ScatterBaseDescription.$t = markType(ScatterBaseDescription, 'ScatterBaseDescription', MarkerSeriesDescription.$);
    return ScatterBaseDescription;
}(MarkerSeriesDescription));
export { ScatterBaseDescription };
