// igniteui-react-core/esm5/lib/ScatterAreaSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ScatterTriangulationSeriesDescription } from "./ScatterTriangulationSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ScatterAreaSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterAreaSeriesDescription, _super);
    function ScatterAreaSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fu = null;
        _this.fz = null;
        _this.ft = null;
        _this.fx = null;
        _this.fy = null;
        return _this;
    }
    ScatterAreaSeriesDescription.prototype.get_type = function () {
        return "ScatterAreaSeries";
    };
    Object.defineProperty(ScatterAreaSeriesDescription.prototype, "colorScale", {
        get: function () {
            return this.fu;
        },
        set: function (a) {
            this.fu = a;
            this.g("ColorScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterAreaSeriesDescription.prototype, "colorMemberPath", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("ColorMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterAreaSeriesDescription.prototype, "actualColorScale", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("ActualColorScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterAreaSeriesDescription.prototype, "colorMemberAsLegendLabel", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("ColorMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterAreaSeriesDescription.prototype, "colorMemberAsLegendUnit", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            this.fy = a;
            this.g("ColorMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    ScatterAreaSeriesDescription.$t = markType(ScatterAreaSeriesDescription, 'ScatterAreaSeriesDescription', ScatterTriangulationSeriesDescription.$);
    return ScatterAreaSeriesDescription;
}(ScatterTriangulationSeriesDescription));
export { ScatterAreaSeriesDescription };
