// igniteui-react-core/esm5/lib/ScatterTriangulationSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ScatterTriangulationSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterTriangulationSeriesDescription, _super);
    function ScatterTriangulationSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fb = null;
        _this.ff = null;
        _this.e8 = null;
        _this.fc = null;
        _this.e3 = null;
        _this.e4 = null;
        _this.e5 = null;
        _this.e6 = null;
        _this.e9 = null;
        _this.fd = null;
        _this.fa = null;
        _this.fe = null;
        _this.e7 = null;
        return _this;
    }
    ScatterTriangulationSeriesDescription.prototype.get_type = function () {
        return "ScatterTriangulationSeries";
    };
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "xMemberPath", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("XMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "yMemberPath", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            this.ff = a;
            this.g("YMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "xAxisRef", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("XAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "yAxisRef", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("YAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "trianglesSourceRef", {
        get: function () {
            return this.e3;
        },
        set: function (a) {
            this.e3 = a;
            this.g("TrianglesSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "triangleVertexMemberPath1", {
        get: function () {
            return this.e4;
        },
        set: function (a) {
            this.e4 = a;
            this.g("TriangleVertexMemberPath1");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "triangleVertexMemberPath2", {
        get: function () {
            return this.e5;
        },
        set: function (a) {
            this.e5 = a;
            this.g("TriangleVertexMemberPath2");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "triangleVertexMemberPath3", {
        get: function () {
            return this.e6;
        },
        set: function (a) {
            this.e6 = a;
            this.g("TriangleVertexMemberPath3");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "xMemberAsLegendLabel", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("XMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "yMemberAsLegendLabel", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("YMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "xMemberAsLegendUnit", {
        get: function () {
            return this.fa;
        },
        set: function (a) {
            this.fa = a;
            this.g("XMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "yMemberAsLegendUnit", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("YMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterTriangulationSeriesDescription.prototype, "triangulationStatusChangedRef", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("TriangulationStatusChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    ScatterTriangulationSeriesDescription.$t = markType(ScatterTriangulationSeriesDescription, 'ScatterTriangulationSeriesDescription', SeriesDescription.$);
    return ScatterTriangulationSeriesDescription;
}(SeriesDescription));
export { ScatterTriangulationSeriesDescription };
