// igniteui-react-core/esm5/lib/ScatterPolygonSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ShapeSeriesBaseDescription } from "./ShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ScatterPolygonSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolygonSeriesDescription, _super);
    function ScatterPolygonSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.f1 = null;
        _this.ge = null;
        _this.gd = null;
        _this.f3 = 0;
        _this.f7 = null;
        _this.f8 = null;
        _this.f5 = null;
        _this.gb = null;
        _this.f6 = null;
        _this.f9 = null;
        _this.gc = null;
        _this.ga = null;
        return _this;
    }
    ScatterPolygonSeriesDescription.prototype.get_type = function () {
        return "ScatterPolygonSeries";
    };
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "shapeStyle", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("ShapeStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerType", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("MarkerType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerTemplateRef", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("MarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerThickness", {
        get: function () {
            return this.f3;
        },
        set: function (a) {
            this.f3 = a;
            this.g("MarkerThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "actualMarkerTemplateRef", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("ActualMarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerBrush", {
        get: function () {
            return this.f8;
        },
        set: function (a) {
            this.f8 = a;
            this.g("MarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "actualMarkerBrush", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("ActualMarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerOutline", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("MarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "actualMarkerOutline", {
        get: function () {
            return this.f6;
        },
        set: function (a) {
            this.f6 = a;
            this.g("ActualMarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerCollisionAvoidance", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("MarkerCollisionAvoidance");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerOutlineMode", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("MarkerOutlineMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerFillMode", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("MarkerFillMode");
        },
        enumerable: false,
        configurable: true
    });
    ScatterPolygonSeriesDescription.$t = markType(ScatterPolygonSeriesDescription, 'ScatterPolygonSeriesDescription', ShapeSeriesBaseDescription.$);
    return ScatterPolygonSeriesDescription;
}(ShapeSeriesBaseDescription));
export { ScatterPolygonSeriesDescription };
