// igniteui-react-core/esm5/lib/SchedulableRenderWrapper.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, typeCast, markType } from "./type";
import { ISchedulableRender_$type } from "./ISchedulableRender";
/**
 * @hidden
 */
var SchedulableRenderWrapper = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SchedulableRenderWrapper, _super);
    function SchedulableRenderWrapper(a) {
        var _this = _super.call(this) || this;
        _this.a = null;
        _this.a = a;
        return _this;
    }
    Object.defineProperty(SchedulableRenderWrapper.prototype, "b", {
        get: function () {
            return this.a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SchedulableRenderWrapper.prototype, "d", {
        get: function () {
            var a = this.b;
            return a != null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SchedulableRenderWrapper.prototype, "isDirty", {
        get: function () {
            var a = this.b;
            return a != null ? a.isDirty : false;
        },
        set: function (a) {
            var b = this.b;
            if (b != null) {
                b.isDirty = a;
            }
        },
        enumerable: false,
        configurable: true
    });
    SchedulableRenderWrapper.prototype.undirty = function (a) {
        var b = this.b;
        if (b != null) {
            b.undirty(a);
        }
    };
    Object.defineProperty(SchedulableRenderWrapper.prototype, "index", {
        get: function () {
            var a = this.b;
            return a != null ? a.index : 0;
        },
        enumerable: false,
        configurable: true
    });
    SchedulableRenderWrapper.prototype.postRender = function () {
        var a = this.b;
        if (a != null) {
            a.postRender();
        }
    };
    SchedulableRenderWrapper.prototype.preRender = function () {
        var a = this.b;
        if (a != null) {
            a.preRender();
        }
    };
    SchedulableRenderWrapper.prototype.isValid = function () {
        var a = this.b;
        if (a != null) {
            a.isValid();
        }
        return false;
    };
    SchedulableRenderWrapper.prototype.equals = function (a) {
        var b = typeCast(SchedulableRenderWrapper.$, a);
        if (b == null) {
            return _super.prototype.equals.call(this, a);
        }
        var c = b.b;
        var d = this.b;
        return Base.equalsStatic(c, d);
    };
    SchedulableRenderWrapper.$t = markType(SchedulableRenderWrapper, 'SchedulableRenderWrapper', Base.$, [ISchedulableRender_$type]);
    return SchedulableRenderWrapper;
}(Base));
export { SchedulableRenderWrapper };
