// igniteui-react-core/esm5/lib/SectionHeaderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var SectionHeaderDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SectionHeaderDescription, _super);
    function SectionHeaderDescription() {
        var _this = _super.call(this) || this;
        _this.cz = null;
        _this.cy = null;
        _this.cr = 0;
        _this.ct = 0;
        _this.cs = 0;
        _this.cq = 0;
        _this.co = false;
        _this.c0 = null;
        return _this;
    }
    SectionHeaderDescription.prototype.get_type = function () {
        return "SectionHeader";
    };
    Object.defineProperty(SectionHeaderDescription.prototype, "selectedBackground", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("SelectedBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SectionHeaderDescription.prototype, "actualSelectedBackground", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("ActualSelectedBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SectionHeaderDescription.prototype, "paddingLeft", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("PaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SectionHeaderDescription.prototype, "paddingTop", {
        get: function () {
            return this.ct;
        },
        set: function (a) {
            this.ct = a;
            this.g("PaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SectionHeaderDescription.prototype, "paddingRight", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("PaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SectionHeaderDescription.prototype, "paddingBottom", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("PaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SectionHeaderDescription.prototype, "isCollapsable", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.g("IsCollapsable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SectionHeaderDescription.prototype, "summaryDisplayMode", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("SummaryDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    SectionHeaderDescription.$t = markType(SectionHeaderDescription, 'SectionHeaderDescription', DefinitionBaseDescription.$);
    return SectionHeaderDescription;
}(DefinitionBaseDescription));
export { SectionHeaderDescription };
