// igniteui-react-core/esm5/lib/SeriesLayerPropertyOverlayDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var SeriesLayerPropertyOverlayDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SeriesLayerPropertyOverlayDescription, _super);
    function SeriesLayerPropertyOverlayDescription() {
        var _this = _super.call(this) || this;
        _this.i = false;
        _this.h = false;
        _this.m = null;
        _this.l = null;
        _this.o = null;
        _this.n = null;
        _this.p = null;
        return _this;
    }
    SeriesLayerPropertyOverlayDescription.prototype.get_type = function () {
        return "SeriesLayerPropertyOverlay";
    };
    Object.defineProperty(SeriesLayerPropertyOverlayDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesLayerPropertyOverlayDescription.prototype, "isSourceOverlay", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("IsSourceOverlay");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesLayerPropertyOverlayDescription.prototype, "isAlwaysApplied", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("IsAlwaysApplied");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesLayerPropertyOverlayDescription.prototype, "propertyName", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("PropertyName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesLayerPropertyOverlayDescription.prototype, "currentValuePropertyName", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("CurrentValuePropertyName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesLayerPropertyOverlayDescription.prototype, "valueRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("ValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesLayerPropertyOverlayDescription.prototype, "propertyUpdatedRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("PropertyUpdatedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesLayerPropertyOverlayDescription.prototype, "valueResolvingRef", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("ValueResolvingRef");
        },
        enumerable: false,
        configurable: true
    });
    SeriesLayerPropertyOverlayDescription.$t = markType(SeriesLayerPropertyOverlayDescription, 'SeriesLayerPropertyOverlayDescription', Description.$);
    return SeriesLayerPropertyOverlayDescription;
}(Description));
export { SeriesLayerPropertyOverlayDescription };
