// igniteui-react-core/esm5/lib/SeriesViewerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var SeriesViewerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SeriesViewerDescription, _super);
    function SeriesViewerDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.h = null;
        _this.l = null;
        _this.j = null;
        _this.k = null;
        _this.eo = null;
        _this.v = null;
        _this.en = null;
        _this.af = false;
        _this.am = false;
        _this.ak = false;
        _this.ap = false;
        _this.ec = null;
        _this.e2 = null;
        _this.e0 = null;
        _this.d8 = null;
        _this.ef = null;
        _this.ft = null;
        _this.eh = null;
        _this.fv = null;
        _this.ee = null;
        _this.eg = null;
        _this.bk = 0;
        _this.bj = 0;
        _this.bg = 0;
        _this.bh = 0;
        _this.bi = 0;
        _this.bm = 0;
        _this.bl = 0;
        _this.ag = false;
        _this.fs = null;
        _this.fu = null;
        _this.b5 = 0;
        _this.b4 = 0;
        _this.b2 = 0;
        _this.b8 = 0;
        _this.b3 = 0;
        _this.b7 = 0;
        _this.b6 = 0;
        _this.as = false;
        _this.eb = null;
        _this.ek = null;
        _this.em = null;
        _this.el = null;
        _this.y = null;
        _this.aa = null;
        _this.ar = false;
        _this.al = false;
        _this.ds = 0;
        _this.dp = 0;
        _this.dr = 0;
        _this.dn = 0;
        _this.dq = 0;
        _this.ai = false;
        _this.d2 = null;
        _this.dy = null;
        _this.aj = false;
        _this.fz = null;
        _this.cc = 0;
        _this.cb = 0;
        _this.bc = 0;
        _this.bb = 0;
        _this.ce = 0;
        _this.cd = 0;
        _this.fj = null;
        _this.d3 = null;
        _this.ed = null;
        _this.fr = null;
        _this.f0 = null;
        _this.eq = null;
        _this.an = false;
        _this.d4 = null;
        _this.ej = null;
        _this.ey = null;
        _this.d5 = null;
        _this.ep = null;
        _this.e3 = null;
        _this.z = null;
        _this.ea = null;
        _this.m = null;
        _this.i = null;
        _this.b9 = 0;
        _this.ca = 0;
        _this.e1 = null;
        _this.d6 = null;
        _this.d1 = null;
        _this.fo = null;
        _this.fg = null;
        _this.fq = null;
        _this.fi = null;
        _this.fp = null;
        _this.fh = null;
        _this.b0 = 0;
        _this.by = 0;
        _this.bz = 0;
        _this.bw = 0;
        _this.bv = 0;
        _this.bt = 0;
        _this.bu = 0;
        _this.br = 0;
        _this.fc = null;
        _this.b1 = 0;
        _this.bo = 0;
        _this.bq = 0;
        _this.bf = 0;
        _this.be = 0;
        _this.bd = 0;
        _this.ah = false;
        _this.bp = 0;
        _this.bn = 0;
        _this.a8 = 0;
        _this.a7 = 0;
        _this.ao = false;
        _this.x = null;
        _this.a9 = 0;
        _this.ba = 0;
        _this.eu = null;
        _this.ev = null;
        _this.er = null;
        _this.es = null;
        _this.et = null;
        _this.ew = null;
        _this.e5 = null;
        _this.e8 = null;
        _this.e9 = null;
        _this.e4 = null;
        _this.fa = null;
        _this.e6 = null;
        _this.e7 = null;
        _this.fx = null;
        _this.fw = null;
        _this.fy = null;
        _this.fb = null;
        _this.dz = null;
        _this.d9 = null;
        _this.ez = null;
        _this.d7 = null;
        _this.ex = null;
        _this.ei = null;
        _this.d0 = null;
        _this.fl = null;
        _this.bx = 0;
        _this.fn = null;
        _this.fm = null;
        _this.fd = null;
        _this.bs = 0;
        _this.ff = null;
        _this.fe = null;
        _this.aq = false;
        _this.fk = null;
        return _this;
    }
    SeriesViewerDescription.prototype.get_type = function () {
        return "SeriesViewer";
    };
    Object.defineProperty(SeriesViewerDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "series", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Series");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "brushes", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Brushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "outlines", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Outlines");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "markerBrushes", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("MarkerBrushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "markerOutlines", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("MarkerOutlines");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "markerAutomaticBehavior", {
        get: function () {
            return this.eo;
        },
        set: function (a) {
            this.eo = a;
            this.g("MarkerAutomaticBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "crosshairPoint", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("CrosshairPoint");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "legendRef", {
        get: function () {
            return this.en;
        },
        set: function (a) {
            this.en = a;
            this.g("LegendRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "animateSeriesWhenAxisRangeChanges", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("AnimateSeriesWhenAxisRangeChanges");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "seriesValueLayerUseLegend", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("SeriesValueLayerUseLegend");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "isWindowSyncedToVisibleRange", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("IsWindowSyncedToVisibleRange");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "shouldSimulateHoverMoveCrosshairPoint", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("ShouldSimulateHoverMoveCrosshairPoint");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "highlightingMode", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            this.ec = a;
            this.g("HighlightingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "selectionMode", {
        get: function () {
            return this.e2;
        },
        set: function (a) {
            this.e2 = a;
            this.g("SelectionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "selectionBehavior", {
        get: function () {
            return this.e0;
        },
        set: function (a) {
            this.e0 = a;
            this.g("SelectionBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "focusMode", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.g("FocusMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarMode", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            this.ef = a;
            this.g("HorizontalViewScrollbarMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarMode", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("VerticalViewScrollbarMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarPosition", {
        get: function () {
            return this.eh;
        },
        set: function (a) {
            this.eh = a;
            this.g("HorizontalViewScrollbarPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarPosition", {
        get: function () {
            return this.fv;
        },
        set: function (a) {
            this.fv = a;
            this.g("VerticalViewScrollbarPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarFill", {
        get: function () {
            return this.ee;
        },
        set: function (a) {
            this.ee = a;
            this.g("HorizontalViewScrollbarFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarOutline", {
        get: function () {
            return this.eg;
        },
        set: function (a) {
            this.eg = a;
            this.g("HorizontalViewScrollbarOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarStrokeThickness", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("HorizontalViewScrollbarStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarMaxOpacity", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("HorizontalViewScrollbarMaxOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarCornerRadius", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.g("HorizontalViewScrollbarCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarHeight", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.g("HorizontalViewScrollbarHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarInset", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("HorizontalViewScrollbarInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarTrackStartInset", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            this.bm = a;
            this.g("HorizontalViewScrollbarTrackStartInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarTrackEndInset", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.g("HorizontalViewScrollbarTrackEndInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalViewScrollbarShouldAddAutoTrackInsets", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("HorizontalViewScrollbarShouldAddAutoTrackInsets");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarFill", {
        get: function () {
            return this.fs;
        },
        set: function (a) {
            this.fs = a;
            this.g("VerticalViewScrollbarFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarOutline", {
        get: function () {
            return this.fu;
        },
        set: function (a) {
            this.fu = a;
            this.g("VerticalViewScrollbarOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarStrokeThickness", {
        get: function () {
            return this.b5;
        },
        set: function (a) {
            this.b5 = a;
            this.g("VerticalViewScrollbarStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarMaxOpacity", {
        get: function () {
            return this.b4;
        },
        set: function (a) {
            this.b4 = a;
            this.g("VerticalViewScrollbarMaxOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarCornerRadius", {
        get: function () {
            return this.b2;
        },
        set: function (a) {
            this.b2 = a;
            this.g("VerticalViewScrollbarCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarWidth", {
        get: function () {
            return this.b8;
        },
        set: function (a) {
            this.b8 = a;
            this.g("VerticalViewScrollbarWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarInset", {
        get: function () {
            return this.b3;
        },
        set: function (a) {
            this.b3 = a;
            this.g("VerticalViewScrollbarInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarTrackStartInset", {
        get: function () {
            return this.b7;
        },
        set: function (a) {
            this.b7 = a;
            this.g("VerticalViewScrollbarTrackStartInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarTrackEndInset", {
        get: function () {
            return this.b6;
        },
        set: function (a) {
            this.b6 = a;
            this.g("VerticalViewScrollbarTrackEndInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalViewScrollbarShouldAddAutoTrackInsets", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("VerticalViewScrollbarShouldAddAutoTrackInsets");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "highlightingBehavior", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("HighlightingBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "legendHighlightingMode", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.g("LegendHighlightingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "legendItemBadgeShape", {
        get: function () {
            return this.em;
        },
        set: function (a) {
            this.em = a;
            this.g("LegendItemBadgeShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "legendItemBadgeMode", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            this.el = a;
            this.g("LegendItemBadgeMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "effectiveViewport", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("EffectiveViewport");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "windowRect", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("WindowRect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "useTiledZooming", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("UseTiledZooming");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "preferHigherResolutionTiles", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("PreferHigherResolutionTiles");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "zoomTileCacheSize", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("ZoomTileCacheSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "highlightingTransitionDuration", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("HighlightingTransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "selectionTransitionDuration", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("SelectionTransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "focusTransitionDuration", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.g("FocusTransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "scrollbarsAnimationDuration", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("ScrollbarsAnimationDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "isPagePanningAllowed", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("IsPagePanningAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "contentHitTestMode", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            this.d2 = a;
            this.g("ContentHitTestMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "actualContentHitTestMode", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("ActualContentHitTestMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "isSurfaceInteractionDisabled", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("IsSurfaceInteractionDisabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "windowResponse", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("WindowResponse");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "windowRectMinWidth", {
        get: function () {
            return this.cc;
        },
        set: function (a) {
            this.cc = a;
            this.g("WindowRectMinWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "windowRectMinHeight", {
        get: function () {
            return this.cb;
        },
        set: function (a) {
            this.cb = a;
            this.g("WindowRectMinHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "actualWindowRectMinWidth", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("ActualWindowRectMinWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "actualWindowRectMinHeight", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("ActualWindowRectMinHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "windowSizeMinWidth", {
        get: function () {
            return this.ce;
        },
        set: function (a) {
            this.ce = a;
            this.g("WindowSizeMinWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "windowSizeMinHeight", {
        get: function () {
            return this.cd;
        },
        set: function (a) {
            this.cd = a;
            this.g("WindowSizeMinHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "syncChannel", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("SyncChannel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "crosshairVisibility", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.g("CrosshairVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "horizontalCrosshairBrush", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            this.ed = a;
            this.g("HorizontalCrosshairBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "verticalCrosshairBrush", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("VerticalCrosshairBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "zoomCoercionMode", {
        get: function () {
            return this.f0;
        },
        set: function (a) {
            this.f0 = a;
            this.g("ZoomCoercionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "plotAreaBackground", {
        get: function () {
            return this.eq;
        },
        set: function (a) {
            this.eq = a;
            this.g("PlotAreaBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "shouldMatchZOrderToSeriesOrder", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("ShouldMatchZOrderToSeriesOrder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "defaultInteraction", {
        get: function () {
            return this.d4;
        },
        set: function (a) {
            this.d4 = a;
            this.g("DefaultInteraction");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "interactionOverride", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.g("InteractionOverride");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "rightButtonDefaultInteraction", {
        get: function () {
            return this.ey;
        },
        set: function (a) {
            this.ey = a;
            this.g("RightButtonDefaultInteraction");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "dragModifier", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("DragModifier");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "panModifier", {
        get: function () {
            return this.ep;
        },
        set: function (a) {
            this.ep = a;
            this.g("PanModifier");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "selectionModifier", {
        get: function () {
            return this.e3;
        },
        set: function (a) {
            this.e3 = a;
            this.g("SelectionModifier");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "previewRect", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("PreviewRect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "highlightedValuesDisplayMode", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.g("HighlightedValuesDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "selectedSeriesItems", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("SelectedSeriesItems");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "focusedSeriesItems", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("FocusedSeriesItems");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "windowPositionHorizontal", {
        get: function () {
            return this.b9;
        },
        set: function (a) {
            this.b9 = a;
            this.g("WindowPositionHorizontal");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "windowPositionVertical", {
        get: function () {
            return this.ca;
        },
        set: function (a) {
            this.ca = a;
            this.g("WindowPositionVertical");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "selectionBrush", {
        get: function () {
            return this.e1;
        },
        set: function (a) {
            this.e1 = a;
            this.g("SelectionBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "focusBrush", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            this.d6 = a;
            this.g("FocusBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "chartTitle", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            this.d1 = a;
            this.g("ChartTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleHorizontalAlignment", {
        get: function () {
            return this.fo;
        },
        set: function (a) {
            this.fo = a;
            this.g("TitleHorizontalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleHorizontalAlignment", {
        get: function () {
            return this.fg;
        },
        set: function (a) {
            this.fg = a;
            this.g("SubtitleHorizontalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleTextStyle", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("TitleTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleTextStyle", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("SubtitleTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleTextColor", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("TitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleTextColor", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("SubtitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleTopMargin", {
        get: function () {
            return this.b0;
        },
        set: function (a) {
            this.b0 = a;
            this.g("TitleTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleLeftMargin", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            this.by = a;
            this.g("TitleLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleRightMargin", {
        get: function () {
            return this.bz;
        },
        set: function (a) {
            this.bz = a;
            this.g("TitleRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleBottomMargin", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("TitleBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleTopMargin", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("SubtitleTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleLeftMargin", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("SubtitleLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleRightMargin", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("SubtitleRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleBottomMargin", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.g("SubtitleBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitle", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("Subtitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "topMargin", {
        get: function () {
            return this.b1;
        },
        set: function (a) {
            this.b1 = a;
            this.g("TopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "leftMargin", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            this.bo = a;
            this.g("LeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "rightMargin", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            this.bq = a;
            this.g("RightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "bottomMargin", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("BottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "autoMarginWidth", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.g("AutoMarginWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "autoMarginHeight", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("AutoMarginHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "isAntiAliasingEnabledDuringInteraction", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("IsAntiAliasingEnabledDuringInteraction");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.g("PixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "interactionPixelScalingRatio", {
        get: function () {
            return this.bn;
        },
        set: function (a) {
            this.bn = a;
            this.g("InteractionPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "actualPixelScalingRatio", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("ActualPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "actualInteractionPixelScalingRatio", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("ActualInteractionPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "shouldPanOnMaximumZoom", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("ShouldPanOnMaximumZoom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "actualWindowRect", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("ActualWindowRect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "actualWindowPositionHorizontal", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("ActualWindowPositionHorizontal");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "actualWindowPositionVertical", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("ActualWindowPositionVertical");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "plotAreaMouseLeftButtonDownRef", {
        get: function () {
            return this.eu;
        },
        set: function (a) {
            this.eu = a;
            this.g("PlotAreaMouseLeftButtonDownRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "plotAreaMouseLeftButtonUpRef", {
        get: function () {
            return this.ev;
        },
        set: function (a) {
            this.ev = a;
            this.g("PlotAreaMouseLeftButtonUpRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "plotAreaClickedRef", {
        get: function () {
            return this.er;
        },
        set: function (a) {
            this.er = a;
            this.g("PlotAreaClickedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "plotAreaMouseEnterRef", {
        get: function () {
            return this.es;
        },
        set: function (a) {
            this.es = a;
            this.g("PlotAreaMouseEnterRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "plotAreaMouseLeaveRef", {
        get: function () {
            return this.et;
        },
        set: function (a) {
            this.et = a;
            this.g("PlotAreaMouseLeaveRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "plotAreaMouseOverRef", {
        get: function () {
            return this.ew;
        },
        set: function (a) {
            this.ew = a;
            this.g("PlotAreaMouseOverRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "seriesCursorMouseMoveRef", {
        get: function () {
            return this.e5;
        },
        set: function (a) {
            this.e5 = a;
            this.g("SeriesCursorMouseMoveRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "seriesMouseLeftButtonDownRef", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("SeriesMouseLeftButtonDownRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "seriesMouseLeftButtonUpRef", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("SeriesMouseLeftButtonUpRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "seriesClickRef", {
        get: function () {
            return this.e4;
        },
        set: function (a) {
            this.e4 = a;
            this.g("SeriesClickRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "seriesMouseMoveRef", {
        get: function () {
            return this.fa;
        },
        set: function (a) {
            this.fa = a;
            this.g("SeriesMouseMoveRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "seriesMouseEnterRef", {
        get: function () {
            return this.e6;
        },
        set: function (a) {
            this.e6 = a;
            this.g("SeriesMouseEnterRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "seriesMouseLeaveRef", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("SeriesMouseLeaveRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "viewerManipulationStartingRef", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("ViewerManipulationStartingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "viewerManipulationEndingRef", {
        get: function () {
            return this.fw;
        },
        set: function (a) {
            this.fw = a;
            this.g("ViewerManipulationEndingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "windowRectChangedRef", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            this.fy = a;
            this.g("WindowRectChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "sizeChangedRef", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("SizeChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "actualWindowRectChangedRef", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.g("ActualWindowRectChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "gridAreaRectChangedRef", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("GridAreaRectChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "selectedSeriesItemsChangedRef", {
        get: function () {
            return this.ez;
        },
        set: function (a) {
            this.ez = a;
            this.g("SelectedSeriesItemsChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "focusedSeriesItemsChangedRef", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("FocusedSeriesItemsChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "refreshCompletedRef", {
        get: function () {
            return this.ex;
        },
        set: function (a) {
            this.ex = a;
            this.g("RefreshCompletedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "imageCapturedRef", {
        get: function () {
            return this.ei;
        },
        set: function (a) {
            this.ei = a;
            this.g("ImageCapturedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "background", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleFontFamily", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("TitleFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleFontSize", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("TitleFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleFontWeight", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("TitleFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "titleFontStyle", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("TitleFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleFontFamily", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("SubtitleFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleFontSize", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("SubtitleFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleFontWeight", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            this.ff = a;
            this.g("SubtitleFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "subtitleFontStyle", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("SubtitleFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "synchronizeHorizontally", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("SynchronizeHorizontally");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesViewerDescription.prototype, "synchronizeVertically", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("SynchronizeVertically");
        },
        enumerable: false,
        configurable: true
    });
    SeriesViewerDescription.$t = markType(SeriesViewerDescription, 'SeriesViewerDescription', Description.$);
    return SeriesViewerDescription;
}(Description));
export { SeriesViewerDescription };
