// igniteui-react-core/esm5/lib/SeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var SeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SeriesDescription, _super);
    function SeriesDescription() {
        var _this = _super.call(this) || this;
        _this.cq = null;
        _this.cx = null;
        _this.l = null;
        _this.cr = null;
        _this.cu = null;
        _this.c5 = null;
        _this.c4 = null;
        _this.c2 = null;
        _this.ci = null;
        _this.ch = null;
        _this.cg = null;
        _this.c1 = null;
        _this.c0 = null;
        _this.c3 = null;
        _this.cs = null;
        _this.b3 = 0;
        _this.dg = null;
        _this.dh = null;
        _this.dk = null;
        _this.b4 = 0;
        _this.az = 0;
        _this.a8 = 0;
        _this.bg = 0;
        _this.bd = 0;
        _this.be = 0;
        _this.bf = 0;
        _this.cw = null;
        _this.cv = null;
        _this.co = null;
        _this.cb = null;
        _this.dc = null;
        _this.ct = null;
        _this.ck = null;
        _this.cc = null;
        _this.db = null;
        _this.c8 = null;
        _this.cj = null;
        _this.c7 = null;
        _this.cy = null;
        _this.aa = false;
        _this.bc = 0;
        _this.a0 = 0;
        _this.k = null;
        _this.ce = null;
        _this.cl = null;
        _this.cd = null;
        _this.aw = 0;
        _this.a3 = 0;
        _this.ax = 0;
        _this.a4 = 0;
        _this.u = false;
        _this.v = false;
        _this.y = false;
        _this.af = false;
        _this.ad = false;
        _this.ac = false;
        _this.ab = false;
        _this.x = false;
        _this.a9 = 0;
        _this.dd = null;
        _this.ag = false;
        _this.ba = 0;
        _this.bb = 0;
        _this.a1 = 0;
        _this.av = 0;
        _this.a5 = 0;
        _this.ay = 0;
        _this.w = false;
        _this.ae = false;
        _this.dn = null;
        _this.c9 = null;
        _this.b5 = 0;
        _this.b6 = 0;
        _this.di = null;
        _this.dl = null;
        _this.c6 = null;
        _this.cm = null;
        _this.i = null;
        _this.cn = null;
        _this.j = null;
        _this.z = false;
        _this.cp = null;
        _this.t = false;
        _this.cz = null;
        _this.cf = null;
        _this.a2 = 0;
        _this.a7 = 0;
        _this.m = null;
        _this.h = null;
        _this.a6 = 0;
        _this.dm = null;
        _this.dj = null;
        _this.da = null;
        _this.df = null;
        _this.de = null;
        return _this;
    }
    SeriesDescription.prototype.get_type = function () {
        return "Series";
    };
    Object.defineProperty(SeriesDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "dataLegendGroup", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("DataLegendGroup");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "highlightedValuesDataLegendGroup", {
        get: function () {
            return this.cx;
        },
        set: function (a) {
            this.cx = a;
            this.g("HighlightedValuesDataLegendGroup");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "highlightedValuesExtraPropertyOverlays", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("HighlightedValuesExtraPropertyOverlays");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "highlightedDataSourceRef", {
        get: function () {
            return this.cu;
        },
        set: function (a) {
            this.cu = a;
            this.g("HighlightedDataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "legendRef", {
        get: function () {
            return this.c5;
        },
        set: function (a) {
            this.c5 = a;
            this.g("LegendRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "legendItemVisibility", {
        get: function () {
            return this.c4;
        },
        set: function (a) {
            this.c4 = a;
            this.g("LegendItemVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "legendItemBadgeTemplateRef", {
        get: function () {
            return this.c2;
        },
        set: function (a) {
            this.c2 = a;
            this.g("LegendItemBadgeTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualLegendItemBadgeTemplateRef", {
        get: function () {
            return this.ci;
        },
        set: function (a) {
            this.ci = a;
            this.g("ActualLegendItemBadgeTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualLegendItemBadgeOutline", {
        get: function () {
            return this.ch;
        },
        set: function (a) {
            this.ch = a;
            this.g("ActualLegendItemBadgeOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualLegendItemBadgeBrush", {
        get: function () {
            return this.cg;
        },
        set: function (a) {
            this.cg = a;
            this.g("ActualLegendItemBadgeBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "legendItemBadgeShape", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.g("LegendItemBadgeShape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "legendItemBadgeMode", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("LegendItemBadgeMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "legendItemTemplateRef", {
        get: function () {
            return this.c3;
        },
        set: function (a) {
            this.c3 = a;
            this.g("LegendItemTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "discreteLegendItemTemplateRef", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("DiscreteLegendItemTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "index", {
        get: function () {
            return this.b3;
        },
        set: function (a) {
            this.b3 = a;
            this.g("Index");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "transitionEasingFunctionRef", {
        get: function () {
            return this.dg;
        },
        set: function (a) {
            this.dg = a;
            this.g("TransitionEasingFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "transitionInEasingFunctionRef", {
        get: function () {
            return this.dh;
        },
        set: function (a) {
            this.dh = a;
            this.g("TransitionInEasingFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "transitionOutEasingFunctionRef", {
        get: function () {
            return this.dk;
        },
        set: function (a) {
            this.dk = a;
            this.g("TransitionOutEasingFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "transitionDuration", {
        get: function () {
            return this.b4;
        },
        set: function (a) {
            this.b4 = a;
            this.g("TransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualResolution", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("ActualResolution");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "resolution", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("Resolution");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "visibleRangeMarginTop", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.g("VisibleRangeMarginTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "visibleRangeMarginBottom", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("VisibleRangeMarginBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "visibleRangeMarginLeft", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.g("VisibleRangeMarginLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "visibleRangeMarginRight", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("VisibleRangeMarginRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "highlightedTitleSuffix", {
        get: function () {
            return this.cw;
        },
        set: function (a) {
            this.cw = a;
            this.g("HighlightedTitleSuffix");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "highlightedLegendItemVisibility", {
        get: function () {
            return this.cv;
        },
        set: function (a) {
            this.cv = a;
            this.g("HighlightedLegendItemVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "brush", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.g("Brush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualBrush", {
        get: function () {
            return this.cb;
        },
        set: function (a) {
            this.cb = a;
            this.g("ActualBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "selectionBrush", {
        get: function () {
            return this.dc;
        },
        set: function (a) {
            this.dc = a;
            this.g("SelectionBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "focusBrush", {
        get: function () {
            return this.ct;
        },
        set: function (a) {
            this.ct = a;
            this.g("FocusBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualSelectionBrush", {
        get: function () {
            return this.ck;
        },
        set: function (a) {
            this.ck = a;
            this.g("ActualSelectionBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualFocusBrush", {
        get: function () {
            return this.cc;
        },
        set: function (a) {
            this.cc = a;
            this.g("ActualFocusBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "safeActualBrush", {
        get: function () {
            return this.db;
        },
        set: function (a) {
            this.db = a;
            this.g("SafeActualBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "outline", {
        get: function () {
            return this.c8;
        },
        set: function (a) {
            this.c8 = a;
            this.g("Outline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualOutline", {
        get: function () {
            return this.cj;
        },
        set: function (a) {
            this.cj = a;
            this.g("ActualOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "lineJoin", {
        get: function () {
            return this.c7;
        },
        set: function (a) {
            this.c7 = a;
            this.g("LineJoin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "highlightedValuesDisplayMode", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("HighlightedValuesDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "shouldAnimateOnDataSourceSwap", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("ShouldAnimateOnDataSourceSwap");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "thickness", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("Thickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualThickness", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("ActualThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "dashArray", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("DashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualHighlightingMode", {
        get: function () {
            return this.ce;
        },
        set: function (a) {
            this.ce = a;
            this.g("ActualHighlightingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualSelectionMode", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            this.cl = a;
            this.g("ActualSelectionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualFocusMode", {
        get: function () {
            return this.cd;
        },
        set: function (a) {
            this.cd = a;
            this.g("ActualFocusMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualHighlightedValuesFadeOpacity", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("ActualHighlightedValuesFadeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "highlightedValuesFadeOpacity", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("HighlightedValuesFadeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualHighlightingFadeOpacity", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("ActualHighlightingFadeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "highlightingFadeOpacity", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("HighlightingFadeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "isActualLegendFinancial", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("IsActualLegendFinancial");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "isComponentHighlightingModeIgnored", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("IsComponentHighlightingModeIgnored");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "isHighlightingEnabled", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("IsHighlightingEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "useItemWiseColors", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("UseItemWiseColors");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "shouldShiftOpacityForSafeActualBrush", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("ShouldShiftOpacityForSafeActualBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "shouldRemoveHighlightedDataOnLayerHidden", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("ShouldRemoveHighlightedDataOnLayerHidden");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "shouldHideAutoCallouts", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("ShouldHideAutoCallouts");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "isDropShadowEnabled", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("IsDropShadowEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "shadowBlur", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("ShadowBlur");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "shadowColor", {
        get: function () {
            return this.dd;
        },
        set: function (a) {
            this.dd = a;
            this.g("ShadowColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "useSingleShadow", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("UseSingleShadow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "shadowOffsetX", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("ShadowOffsetX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "shadowOffsetY", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("ShadowOffsetY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "areaFillOpacity", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("AreaFillOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualAreaFillOpacity", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("ActualAreaFillOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "markerFillOpacity", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("MarkerFillOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualMarkerFillOpacity", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("ActualMarkerFillOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "isDefaultToolTipSelected", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("IsDefaultToolTipSelected");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "showDefaultTooltip", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("ShowDefaultTooltip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "visibleRangeMode", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.g("VisibleRangeMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "outlineMode", {
        get: function () {
            return this.c9;
        },
        set: function (a) {
            this.c9 = a;
            this.g("OutlineMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "transitionInDuration", {
        get: function () {
            return this.b5;
        },
        set: function (a) {
            this.b5 = a;
            this.g("TransitionInDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "transitionOutDuration", {
        get: function () {
            return this.b6;
        },
        set: function (a) {
            this.b6 = a;
            this.g("TransitionOutDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "transitionInSpeedType", {
        get: function () {
            return this.di;
        },
        set: function (a) {
            this.di = a;
            this.g("TransitionInSpeedType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "transitionOutSpeedType", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            this.dl = a;
            this.g("TransitionOutSpeedType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "lineCap", {
        get: function () {
            return this.c6;
        },
        set: function (a) {
            this.c6 = a;
            this.g("LineCap");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "autoCalloutLabelFormat", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.g("AutoCalloutLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "autoCalloutLabelFormatSpecifiers", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("AutoCalloutLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "autoCalloutValueLabelFormat", {
        get: function () {
            return this.cn;
        },
        set: function (a) {
            this.cn = a;
            this.g("AutoCalloutValueLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "autoCalloutValueLabelFormatSpecifiers", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("AutoCalloutValueLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "mouseOverEnabled", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("MouseOverEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "coercionMethodsRef", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("CoercionMethodsRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "expectFunctions", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("ExpectFunctions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "hitTestMode", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("HitTestMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualHitTestMode", {
        get: function () {
            return this.cf;
        },
        set: function (a) {
            this.cf = a;
            this.g("ActualHitTestMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "finalValue", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("FinalValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "percentChange", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("PercentChange");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "layers", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Layers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "actualLayers", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ActualLayers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "opacity", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("Opacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "visibility", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            this.dm = a;
            this.g("Visibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "transitionOutCompletedRef", {
        get: function () {
            return this.dj;
        },
        set: function (a) {
            this.dj = a;
            this.g("TransitionOutCompletedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "renderRequestedRef", {
        get: function () {
            return this.da;
        },
        set: function (a) {
            this.da = a;
            this.g("RenderRequestedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "tooltipTemplateRef", {
        get: function () {
            return this.df;
        },
        set: function (a) {
            this.df = a;
            this.g("TooltipTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SeriesDescription.prototype, "title", {
        get: function () {
            return this.de;
        },
        set: function (a) {
            this.de = a;
            this.g("Title");
        },
        enumerable: false,
        configurable: true
    });
    SeriesDescription.$t = markType(SeriesDescription, 'SeriesDescription', Description.$);
    return SeriesDescription;
}(Description));
export { SeriesDescription };
