// igniteui-react-core/esm5/lib/ShapeDataSourceDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ShapeDataSourceDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ShapeDataSourceDescription, _super);
    function ShapeDataSourceDescription() {
        var _this = _super.call(this) || this;
        _this.i = null;
        _this.h = null;
        _this.l = false;
        _this.r = null;
        _this.n = null;
        _this.q = null;
        _this.p = null;
        _this.o = null;
        return _this;
    }
    ShapeDataSourceDescription.prototype.get_type = function () {
        return "ShapeDataSource";
    };
    Object.defineProperty(ShapeDataSourceDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeDataSourceDescription.prototype, "worldRect", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("WorldRect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeDataSourceDescription.prototype, "computedWorldRect", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ComputedWorldRect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeDataSourceDescription.prototype, "deferImportCompleted", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("DeferImportCompleted");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeDataSourceDescription.prototype, "shapefileSource", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("ShapefileSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeDataSourceDescription.prototype, "databaseSource", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("DatabaseSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeDataSourceDescription.prototype, "importPendingRef", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("ImportPendingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeDataSourceDescription.prototype, "importCompletedRef", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("ImportCompletedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeDataSourceDescription.prototype, "filterRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("FilterRef");
        },
        enumerable: false,
        configurable: true
    });
    ShapeDataSourceDescription.$t = markType(ShapeDataSourceDescription, 'ShapeDataSourceDescription', Description.$);
    return ShapeDataSourceDescription;
}(Description));
export { ShapeDataSourceDescription };
