// igniteui-react-core/esm5/lib/SizeScaleDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var SizeScaleDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SizeScaleDescription, _super);
    function SizeScaleDescription() {
        var _this = _super.call(this) || this;
        _this.k = 0;
        _this.j = 0;
        _this.m = 0;
        _this.l = 0;
        _this.h = false;
        _this.r = 0;
        _this.t = null;
        return _this;
    }
    SizeScaleDescription.prototype.get_type = function () {
        return "SizeScale";
    };
    Object.defineProperty(SizeScaleDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SizeScaleDescription.prototype, "globalMinimum", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("GlobalMinimum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SizeScaleDescription.prototype, "globalMaximum", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("GlobalMaximum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SizeScaleDescription.prototype, "minimumValue", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("MinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SizeScaleDescription.prototype, "maximumValue", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("MaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SizeScaleDescription.prototype, "isLogarithmic", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("IsLogarithmic");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SizeScaleDescription.prototype, "logarithmBase", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("LogarithmBase");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SizeScaleDescription.prototype, "propertyUpdatedRef", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("PropertyUpdatedRef");
        },
        enumerable: false,
        configurable: true
    });
    SizeScaleDescription.$t = markType(SizeScaleDescription, 'SizeScaleDescription', Description.$);
    return SizeScaleDescription;
}(Description));
export { SizeScaleDescription };
