// igniteui-react-core/esm5/lib/SliceEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var SliceEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SliceEventArgsDescription, _super);
    function SliceEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.k = false;
        _this.j = false;
        _this.h = null;
        _this.n = null;
        return _this;
    }
    SliceEventArgsDescription.prototype.get_type = function () {
        return "SliceEventArgs";
    };
    Object.defineProperty(SliceEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SliceEventArgsDescription.prototype, "isSelected", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("IsSelected");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SliceEventArgsDescription.prototype, "isExploded", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("IsExploded");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SliceEventArgsDescription.prototype, "position", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Position");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SliceEventArgsDescription.prototype, "originalEventRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("OriginalEventRef");
        },
        enumerable: false,
        configurable: true
    });
    SliceEventArgsDescription.$t = markType(SliceEventArgsDescription, 'SliceEventArgsDescription', Description.$);
    return SliceEventArgsDescription;
}(Description));
export { SliceEventArgsDescription };
