// igniteui-react-core/esm5/lib/SortDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, markType } from "./type";
/**
 * @hidden
 */
var SortDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SortDescription, _super);
    function SortDescription() {
        var _this = _super.call(this) || this;
        _this.b = false;
        _this.h = null;
        _this.d = 0;
        _this.g = null;
        return _this;
    }
    Object.defineProperty(SortDescription.prototype, "propertyName", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SortDescription.prototype, "direction", {
        get: function () {
            return this.d;
        },
        set: function (a) {
            this.d = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SortDescription.prototype, "displayName", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
        },
        enumerable: false,
        configurable: true
    });
    SortDescription.prototype.equals = function (a) {
        var b = a;
        if (b == null) {
            return _super.prototype.equals.call(this, a);
        }
        return b.propertyName == this.propertyName && b.direction == this.direction && b.displayName == this.displayName;
    };
    SortDescription.prototype.getHashCode = function () {
        return Base.getHashCodeStatic(this.propertyName) * 17 + this.direction;
    };
    SortDescription.prototype.k = function () {
        this.b = true;
    };
    SortDescription.create = function (a, b, c) {
        if (c === void 0) {
            c = null;
        }
        var d = new SortDescription();
        d.propertyName = a;
        d.direction = b;
        d.displayName = c;
        return d;
    };
    SortDescription.$t = markType(SortDescription, 'SortDescription');
    return SortDescription;
}(Base));
export { SortDescription };
