// igniteui-react-core/esm5/lib/StackedSeriesCreatedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var StackedSeriesCreatedEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StackedSeriesCreatedEventArgsDescription, _super);
    function StackedSeriesCreatedEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.r = null;
        _this.t = null;
        _this.s = null;
        _this.u = null;
        _this.aa = null;
        _this.h = null;
        _this.m = 0;
        _this.ab = null;
        _this.p = 0;
        _this.ac = null;
        _this.w = null;
        _this.x = null;
        _this.j = null;
        _this.y = null;
        _this.z = null;
        _this.l = 0;
        _this.v = null;
        return _this;
    }
    StackedSeriesCreatedEventArgsDescription.prototype.get_type = function () {
        return "StackedSeriesCreatedEventArgs";
    };
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "brush", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Brush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "legendItemTemplateRef", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("LegendItemTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "legendItemBadgeTemplateRef", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("LegendItemBadgeTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "legendItemVisibility", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("LegendItemVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "outline", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("Outline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "dashArray", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("DashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "thickness", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Thickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "titleRef", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("TitleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "transitionDuration", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("TransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "transitionEasingFunctionRef", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("TransitionEasingFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "markerBrush", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("MarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "markerOutline", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("MarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "markerStyle", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("MarkerStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "markerTemplateRef", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("MarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "markerType", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("MarkerType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "markerThickness", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("MarkerThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesCreatedEventArgsDescription.prototype, "lineCap", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("LineCap");
        },
        enumerable: false,
        configurable: true
    });
    StackedSeriesCreatedEventArgsDescription.$t = markType(StackedSeriesCreatedEventArgsDescription, 'StackedSeriesCreatedEventArgsDescription', Description.$);
    return StackedSeriesCreatedEventArgsDescription;
}(Description));
export { StackedSeriesCreatedEventArgsDescription };
