// igniteui-react-core/esm5/lib/StyleShapeEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var StyleShapeEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StyleShapeEventArgsDescription, _super);
    function StyleShapeEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.l = null;
        _this.m = null;
        _this.n = null;
        _this.i = 0;
        _this.h = 0;
        return _this;
    }
    StyleShapeEventArgsDescription.prototype.get_type = function () {
        return "StyleShapeEventArgs";
    };
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "shapeFill", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("ShapeFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "shapeStroke", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ShapeStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "shapeStrokeThickness", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ShapeStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "shapeOpacity", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ShapeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    StyleShapeEventArgsDescription.$t = markType(StyleShapeEventArgsDescription, 'StyleShapeEventArgsDescription', Description.$);
    return StyleShapeEventArgsDescription;
}(Description));
export { StyleShapeEventArgsDescription };
