// igniteui-react-core/esm5/lib/SummaryRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var SummaryRowDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SummaryRowDescription, _super);
    function SummaryRowDescription() {
        var _this = _super.call(this) || this;
        _this.c0 = null;
        _this.cy = null;
        _this.c4 = null;
        _this.cz = null;
        _this.c1 = null;
        _this.co = 0;
        _this.c2 = null;
        _this.c3 = null;
        _this.cr = 0;
        _this.ct = 0;
        _this.cs = 0;
        _this.cq = 0;
        return _this;
    }
    SummaryRowDescription.prototype.get_type = function () {
        return "SummaryRow";
    };
    Object.defineProperty(SummaryRowDescription.prototype, "selectedBackground", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("SelectedBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "actualSelectedBackground", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("ActualSelectedBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "summaryLabelTextColor", {
        get: function () {
            return this.c4;
        },
        set: function (a) {
            this.c4 = a;
            this.g("SummaryLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "actualSummaryLabelTextColor", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("ActualSummaryLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "summaryLabelFontFamily", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.g("SummaryLabelFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "summaryLabelFontSize", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.g("SummaryLabelFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "summaryLabelFontStyle", {
        get: function () {
            return this.c2;
        },
        set: function (a) {
            this.c2 = a;
            this.g("SummaryLabelFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "summaryLabelFontWeight", {
        get: function () {
            return this.c3;
        },
        set: function (a) {
            this.c3 = a;
            this.g("SummaryLabelFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "paddingLeft", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("PaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "paddingTop", {
        get: function () {
            return this.ct;
        },
        set: function (a) {
            this.ct = a;
            this.g("PaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "paddingRight", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("PaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryRowDescription.prototype, "paddingBottom", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("PaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    SummaryRowDescription.$t = markType(SummaryRowDescription, 'SummaryRowDescription', DefinitionBaseDescription.$);
    return SummaryRowDescription;
}(DefinitionBaseDescription));
export { SummaryRowDescription };
