// igniteui-react-core/esm5/lib/SummaryCellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { CellInfoDescription } from "./CellInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var SummaryCellInfoDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SummaryCellInfoDescription, _super);
    function SummaryCellInfoDescription() {
        var _this = _super.call(this) || this;
        _this.fp = null;
        _this.fq = null;
        _this.fu = null;
        _this.fr = null;
        _this.fm = 0;
        _this.fs = null;
        _this.ft = null;
        return _this;
    }
    SummaryCellInfoDescription.prototype.get_type = function () {
        return "SummaryCellInfo";
    };
    Object.defineProperty(SummaryCellInfoDescription.prototype, "resolvedSummaryLabel", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("ResolvedSummaryLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryCellInfoDescription.prototype, "resolvedSummaryValue", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("ResolvedSummaryValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryCellInfoDescription.prototype, "summaryLabelTextColor", {
        get: function () {
            return this.fu;
        },
        set: function (a) {
            this.fu = a;
            this.g("SummaryLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryCellInfoDescription.prototype, "summaryLabelFontFamily", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("SummaryLabelFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryCellInfoDescription.prototype, "summaryLabelFontSize", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("SummaryLabelFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryCellInfoDescription.prototype, "summaryLabelFontStyle", {
        get: function () {
            return this.fs;
        },
        set: function (a) {
            this.fs = a;
            this.g("SummaryLabelFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SummaryCellInfoDescription.prototype, "summaryLabelFontWeight", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("SummaryLabelFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    SummaryCellInfoDescription.$t = markType(SummaryCellInfoDescription, 'SummaryCellInfoDescription', CellInfoDescription.$);
    SummaryCellInfoDescription.__marshalByValue1 = true;
    SummaryCellInfoDescription.__marshalByValueAlias1 = "SummaryCellModel";
    return SummaryCellInfoDescription;
}(CellInfoDescription));
export { SummaryCellInfoDescription };
