// igniteui-react-core/esm5/lib/TestFilterSyntaxVisitor.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { FilterSyntaxVisitor } from "./FilterSyntaxVisitor";
import { StringBuilder } from "./StringBuilder";
import { markType } from "./type";
import { FilterSyntaxPrintingHelper } from "./FilterSyntaxPrintingHelper";
/**
 * @hidden
 */
var TestFilterSyntaxVisitor = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TestFilterSyntaxVisitor, _super);
    function TestFilterSyntaxVisitor() {
        var _this = _super.call(this) || this;
        _this.l = new StringBuilder(0);
        return _this;
    }
    Object.defineProperty(TestFilterSyntaxVisitor.prototype, "k", {
        get: function () {
            return this.l.toString();
        },
        enumerable: false,
        configurable: true
    });
    TestFilterSyntaxVisitor.prototype.c = function (a) {
        this.l.l("[[");
        this.l.l(FilterSyntaxPrintingHelper.b(a.d));
        this.l.l(" ");
        this.b(a.b);
        this.l.l(" ");
        this.b(a.c);
        this.l.l("]]");
    };
    TestFilterSyntaxVisitor.prototype.d = function (a) {
        this.l.l(FilterSyntaxPrintingHelper.a(a));
    };
    TestFilterSyntaxVisitor.prototype.e = function (a) {
        this.b(a.b);
        if (a.c != null) {
            this.l.l(".");
            this.b(a.c);
        }
    };
    TestFilterSyntaxVisitor.prototype.f = function (a) {
        this.b(a.c);
        this.l.l("(");
        if (a.b != null) {
            var b = true;
            var e = a.b;
            for (var d = 0; d < e.length; d++) {
                var c = e[d];
                if (b) {
                    b = false;
                }
                else {
                    this.l.l(", ");
                }
                this.b(c);
            }
        }
        this.l.l(")");
    };
    TestFilterSyntaxVisitor.prototype.g = function (a) {
        this.l.l("null");
    };
    TestFilterSyntaxVisitor.prototype.h = function (a) {
        this.l.l("(");
        this.b(a.b);
        this.l.l(")");
    };
    TestFilterSyntaxVisitor.prototype.i = function (a) {
        this.l.l(FilterSyntaxPrintingHelper.c(a));
    };
    TestFilterSyntaxVisitor.prototype.j = function (a) {
        this.l.l("[[");
        this.l.l(FilterSyntaxPrintingHelper.b(a.c));
        this.l.l(" ");
        this.b(a.b);
        this.l.l("]]");
    };
    TestFilterSyntaxVisitor.$t = markType(TestFilterSyntaxVisitor, 'TestFilterSyntaxVisitor', FilterSyntaxVisitor.$);
    return TestFilterSyntaxVisitor;
}(FilterSyntaxVisitor));
export { TestFilterSyntaxVisitor };
