// igniteui-react-core/esm5/lib/TextHeaderCellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { TextCellInfoDescription } from "./TextCellInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var TextHeaderCellInfoDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TextHeaderCellInfoDescription, _super);
    function TextHeaderCellInfoDescription() {
        var _this = _super.call(this) || this;
        _this.fq = false;
        _this.fw = null;
        _this.ft = null;
        _this.fv = null;
        _this.fu = null;
        return _this;
    }
    TextHeaderCellInfoDescription.prototype.get_type = function () {
        return "TextHeaderCellInfo";
    };
    Object.defineProperty(TextHeaderCellInfoDescription.prototype, "isColumnOptionsEnabled", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("IsColumnOptionsEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextHeaderCellInfoDescription.prototype, "sortIndicatorStyle", {
        get: function () {
            return this.fw;
        },
        set: function (a) {
            this.fw = a;
            this.g("SortIndicatorStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextHeaderCellInfoDescription.prototype, "columnOptionsIconAlignment", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("ColumnOptionsIconAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextHeaderCellInfoDescription.prototype, "columnOptionsIconColor", {
        get: function () {
            return this.fv;
        },
        set: function (a) {
            this.fv = a;
            this.g("ColumnOptionsIconColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextHeaderCellInfoDescription.prototype, "columnOptionsIconBehavior", {
        get: function () {
            return this.fu;
        },
        set: function (a) {
            this.fu = a;
            this.g("ColumnOptionsIconBehavior");
        },
        enumerable: false,
        configurable: true
    });
    TextHeaderCellInfoDescription.$t = markType(TextHeaderCellInfoDescription, 'TextHeaderCellInfoDescription', TextCellInfoDescription.$);
    TextHeaderCellInfoDescription.__marshalByValue2 = true;
    TextHeaderCellInfoDescription.__marshalByValueAlias2 = "TextHeaderCellModel";
    return TextHeaderCellInfoDescription;
}(TextCellInfoDescription));
export { TextHeaderCellInfoDescription };
