// igniteui-react-core/esm5/lib/TextColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DataGridColumnDescription } from "./DataGridColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var TextColumnDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TextColumnDescription, _super);
    function TextColumnDescription() {
        var _this = _super.call(this) || this;
        _this.f7 = null;
        _this.f5 = null;
        _this.f1 = null;
        _this.f6 = null;
        _this.f8 = null;
        return _this;
    }
    TextColumnDescription.prototype.get_type = function () {
        return "TextColumn";
    };
    Object.defineProperty(TextColumnDescription.prototype, "editorType", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("EditorType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextColumnDescription.prototype, "editorDataSourceRef", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("EditorDataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextColumnDescription.prototype, "editorDataSource", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("EditorDataSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextColumnDescription.prototype, "editorTextField", {
        get: function () {
            return this.f6;
        },
        set: function (a) {
            this.f6 = a;
            this.g("EditorTextField");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextColumnDescription.prototype, "editorValueField", {
        get: function () {
            return this.f8;
        },
        set: function (a) {
            this.f8 = a;
            this.g("EditorValueField");
        },
        enumerable: false,
        configurable: true
    });
    TextColumnDescription.$t = markType(TextColumnDescription, 'TextColumnDescription', DataGridColumnDescription.$);
    TextColumnDescription.__canMarshalByValue1 = true;
    TextColumnDescription.__marshalByValueAlias1 = "TextColumn";
    return TextColumnDescription;
}(DataGridColumnDescription));
export { TextColumnDescription };
