// igniteui-react-core/esm5/lib/TimeAxisBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { CategoryAxisBaseDescription } from "./CategoryAxisBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var TimeAxisBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TimeAxisBaseDescription, _super);
    function TimeAxisBaseDescription() {
        var _this = _super.call(this) || this;
        _this.ef = null;
        _this.d3 = false;
        _this.d6 = new Date();
        _this.d5 = new Date();
        _this.d8 = new Date();
        _this.d7 = new Date();
        _this.ee = null;
        _this.ed = null;
        return _this;
    }
    TimeAxisBaseDescription.prototype.get_type = function () {
        return "TimeAxisBase";
    };
    Object.defineProperty(TimeAxisBaseDescription.prototype, "dateTimeMemberPath", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            this.ef = a;
            this.g("DateTimeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisBaseDescription.prototype, "isDataPreSorted", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.g("IsDataPreSorted");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisBaseDescription.prototype, "actualMinimumValue", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            this.d6 = a;
            this.g("ActualMinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisBaseDescription.prototype, "actualMaximumValue", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("ActualMaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisBaseDescription.prototype, "minimumValue", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.g("MinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisBaseDescription.prototype, "maximumValue", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("MaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisBaseDescription.prototype, "actualMinimumValueChangeRef", {
        get: function () {
            return this.ee;
        },
        set: function (a) {
            this.ee = a;
            this.g("ActualMinimumValueChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisBaseDescription.prototype, "actualMaximumValueChangeRef", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            this.ed = a;
            this.g("ActualMaximumValueChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    TimeAxisBaseDescription.$t = markType(TimeAxisBaseDescription, 'TimeAxisBaseDescription', CategoryAxisBaseDescription.$);
    return TimeAxisBaseDescription;
}(CategoryAxisBaseDescription));
export { TimeAxisBaseDescription };
