// igniteui-react-core/esm5/lib/ToolActionCheckboxListDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolActionCheckboxListDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolActionCheckboxListDescription, _super);
    function ToolActionCheckboxListDescription() {
        var _this = _super.call(this) || this;
        _this.dw = null;
        _this.dy = null;
        _this.dt = false;
        _this.dv = null;
        _this.dp = null;
        _this.dq = null;
        _this.dx = null;
        return _this;
    }
    ToolActionCheckboxListDescription.prototype.get_type = function () {
        return "ToolActionCheckboxList";
    };
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "selectedMemberPath", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("SelectedMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "showSelectAll", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("ShowSelectAll");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "dataMemberPath", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.g("DataMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "primaryKey", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("PrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "selectedKeys", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("SelectedKeys");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "indexType", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("IndexType");
        },
        enumerable: false,
        configurable: true
    });
    ToolActionCheckboxListDescription.$t = markType(ToolActionCheckboxListDescription, 'ToolActionCheckboxListDescription', ToolActionDescription.$);
    return ToolActionCheckboxListDescription;
}(ToolActionDescription));
export { ToolActionCheckboxListDescription };
