// igniteui-react-core/esm5/lib/ToolActionRadioDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolActionRadioDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolActionRadioDescription, _super);
    function ToolActionRadioDescription() {
        var _this = _super.call(this) || this;
        _this.dx = null;
        _this.dp = false;
        _this.dq = false;
        _this.d0 = null;
        _this.dv = null;
        _this.dy = null;
        _this.dt = null;
        _this.d1 = null;
        _this.dw = null;
        _this.dz = null;
        _this.du = null;
        return _this;
    }
    ToolActionRadioDescription.prototype.get_type = function () {
        return "ToolActionRadio";
    };
    Object.defineProperty(ToolActionRadioDescription.prototype, "channel", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("Channel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionRadioDescription.prototype, "isChecked", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("IsChecked");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionRadioDescription.prototype, "isManual", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("IsManual");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionRadioDescription.prototype, "uncheckedBackgroundColor", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.g("UncheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionRadioDescription.prototype, "actualUncheckedBackgroundColor", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.g("ActualUncheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionRadioDescription.prototype, "checkedBackgroundColor", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("CheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionRadioDescription.prototype, "actualCheckedBackgroundColor", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("ActualCheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionRadioDescription.prototype, "uncheckedBorderColor", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            this.d1 = a;
            this.g("UncheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionRadioDescription.prototype, "actualUncheckedBorderColor", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.g("ActualUncheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionRadioDescription.prototype, "checkedBorderColor", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.g("CheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionRadioDescription.prototype, "actualCheckedBorderColor", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.g("ActualCheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    ToolActionRadioDescription.$t = markType(ToolActionRadioDescription, 'ToolActionRadioDescription', ToolActionDescription.$);
    return ToolActionRadioDescription;
}(ToolActionDescription));
export { ToolActionRadioDescription };
