// igniteui-react-core/esm5/lib/ToolActionCheckboxListInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ToolActionInfoDescription } from "./ToolActionInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolActionCheckboxListInfoDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolActionCheckboxListInfoDescription, _super);
    function ToolActionCheckboxListInfoDescription() {
        var _this = _super.call(this) || this;
        _this.bj = null;
        _this.bi = null;
        _this.bk = null;
        _this.bf = false;
        _this.bc = null;
        return _this;
    }
    ToolActionCheckboxListInfoDescription.prototype.get_type = function () {
        return "ToolActionCheckboxListInfo";
    };
    Object.defineProperty(ToolActionCheckboxListInfoDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListInfoDescription.prototype, "dataMemberPath", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("DataMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListInfoDescription.prototype, "selectedMemberPath", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("SelectedMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListInfoDescription.prototype, "showSelectAll", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("ShowSelectAll");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListInfoDescription.prototype, "primaryKey", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("PrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    ToolActionCheckboxListInfoDescription.$t = markType(ToolActionCheckboxListInfoDescription, 'ToolActionCheckboxListInfoDescription', ToolActionInfoDescription.$);
    ToolActionCheckboxListInfoDescription.__marshalByValue1 = true;
    ToolActionCheckboxListInfoDescription.__marshalByValueAlias1 = "ToolActionCheckboxListInfo";
    return ToolActionCheckboxListInfoDescription;
}(ToolActionInfoDescription));
export { ToolActionCheckboxListInfoDescription };
