// igniteui-react-core/esm5/lib/ToolbarDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolbarDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolbarDescription, _super);
    function ToolbarDescription() {
        var _this = _super.call(this) || this;
        _this.a1 = null;
        _this.a7 = null;
        _this.j = null;
        _this.h = null;
        _this.i = null;
        _this.bj = null;
        _this.bs = null;
        _this.a0 = null;
        _this.bl = null;
        _this.a9 = null;
        _this.bk = null;
        _this.be = null;
        _this.bt = null;
        _this.a8 = null;
        _this.bb = null;
        _this.bv = null;
        _this.bu = null;
        _this.ad = 0;
        _this.ae = 0;
        _this.bf = null;
        _this.bg = null;
        _this.a2 = null;
        _this.r = 0;
        _this.a3 = null;
        _this.a4 = null;
        _this.bo = null;
        _this.ac = 0;
        _this.bp = null;
        _this.bq = null;
        _this.z = 0;
        _this.aa = 0;
        _this.ab = 0;
        _this.y = 0;
        _this.v = 0;
        _this.w = 0;
        _this.x = 0;
        _this.u = 0;
        _this.n = false;
        _this.o = false;
        _this.at = 0;
        _this.t = 0;
        _this.s = 0;
        _this.bd = null;
        _this.bh = null;
        _this.bw = null;
        _this.a5 = null;
        _this.bx = null;
        _this.a6 = null;
        _this.av = 0;
        _this.br = null;
        _this.ba = null;
        _this.au = 0;
        _this.bi = null;
        _this.bn = null;
        _this.bm = null;
        _this.by = null;
        _this.bc = null;
        _this.az = null;
        return _this;
    }
    ToolbarDescription.prototype.get_type = function () {
        return "Toolbar";
    };
    Object.defineProperty(ToolbarDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "baseTheme", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("BaseTheme");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "density", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("Density");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "autoGeneratedActions", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("AutoGeneratedActions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "actions", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Actions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "actualActions", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ActualActions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "orientation", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("Orientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "targetRef", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("TargetRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "backgroundColor", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("BackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "submenuBackgroundColor", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.g("SubmenuBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "groupHeaderBackgroundColor", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("GroupHeaderBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "separatorBackgroundColor", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("SeparatorBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "hoverBackgroundColor", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.g("HoverBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "textColor", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "disabledTextColor", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("DisabledTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "groupHeaderTextColor", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("GroupHeaderTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "toolTipTextColor", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("ToolTipTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "toolTipBackgroundColor", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("ToolTipBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "toolTipCornerRadius", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("ToolTipCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "toolTipElevation", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("ToolTipElevation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "iconFill", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("IconFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "iconStroke", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.g("IconStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "cellFontFamily", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("CellFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "cellFontSize", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("CellFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "cellFontStyle", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("CellFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "cellFontWeight", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("CellFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "subtitleFontFamily", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            this.bo = a;
            this.g("SubtitleFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "subtitleFontSize", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("SubtitleFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "subtitleFontStyle", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.g("SubtitleFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "subtitleFontWeight", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            this.bq = a;
            this.g("SubtitleFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "separatorVerticalPaddingLeft", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("SeparatorVerticalPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "separatorVerticalPaddingRight", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("SeparatorVerticalPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "separatorVerticalPaddingTop", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("SeparatorVerticalPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "separatorVerticalPaddingBottom", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("SeparatorVerticalPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "separatorHorizontalPaddingLeft", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("SeparatorHorizontalPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "separatorHorizontalPaddingRight", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("SeparatorHorizontalPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "separatorHorizontalPaddingTop", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("SeparatorHorizontalPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "separatorHorizontalPaddingBottom", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("SeparatorHorizontalPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "showOnHover", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ShowOnHover");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "showTooltipOnHover", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("ShowTooltipOnHover");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "dropdownClickBuffer", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("DropdownClickBuffer");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "highlightWidth", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("HighlightWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "highlightRadius", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("HighlightRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "highlightColor", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("HighlightColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "menuArrowStroke", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.g("MenuArrowStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "uncheckedBackgroundColor", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("UncheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "checkedBackgroundColor", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("CheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "uncheckedBorderColor", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("UncheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "checkedBorderColor", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("CheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "rowHeight", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("RowHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "subtitleTextColor", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.g("SubtitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "groupHeaderSubtitleTextColor", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("GroupHeaderSubtitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "dropdownDelay", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("DropdownDelay");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "onCommandRef", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("OnCommandRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "subMenuOpeningRef", {
        get: function () {
            return this.bn;
        },
        set: function (a) {
            this.bn = a;
            this.g("SubMenuOpeningRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "subMenuClosingRef", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            this.bm = a;
            this.g("SubMenuClosingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "width", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            this.by = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "height", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolbarDescription.prototype, "background", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    ToolbarDescription.$t = markType(ToolbarDescription, 'ToolbarDescription', Description.$);
    return ToolbarDescription;
}(Description));
export { ToolbarDescription };
