// igniteui-react-core/esm5/lib/ToolActionComboInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ToolActionInfoDescription } from "./ToolActionInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolActionComboInfoDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolActionComboInfoDescription, _super);
    function ToolActionComboInfoDescription() {
        var _this = _super.call(this) || this;
        _this.be = null;
        _this.bf = null;
        _this.bh = null;
        _this.bg = null;
        return _this;
    }
    ToolActionComboInfoDescription.prototype.get_type = function () {
        return "ToolActionComboInfo";
    };
    Object.defineProperty(ToolActionComboInfoDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionComboInfoDescription.prototype, "displayMemberPath", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("DisplayMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionComboInfoDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionComboInfoDescription.prototype, "selectedValuesRef", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.g("SelectedValuesRef");
        },
        enumerable: false,
        configurable: true
    });
    ToolActionComboInfoDescription.$t = markType(ToolActionComboInfoDescription, 'ToolActionComboInfoDescription', ToolActionInfoDescription.$);
    ToolActionComboInfoDescription.__marshalByValue1 = true;
    ToolActionComboInfoDescription.__marshalByValueAlias1 = "ToolActionComboInfo";
    return ToolActionComboInfoDescription;
}(ToolActionInfoDescription));
export { ToolActionComboInfoDescription };
