// igniteui-react-core/esm5/lib/ToolActionComboDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolActionComboDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolActionComboDescription, _super);
    function ToolActionComboDescription() {
        var _this = _super.call(this) || this;
        _this.dr = null;
        _this.ds = null;
        _this.dt = null;
        _this.dp = null;
        return _this;
    }
    ToolActionComboDescription.prototype.get_type = function () {
        return "ToolActionCombo";
    };
    Object.defineProperty(ToolActionComboDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionComboDescription.prototype, "displayMemberPath", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("DisplayMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionComboDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionComboDescription.prototype, "selectedValues", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("SelectedValues");
        },
        enumerable: false,
        configurable: true
    });
    ToolActionComboDescription.$t = markType(ToolActionComboDescription, 'ToolActionComboDescription', ToolActionDescription.$);
    return ToolActionComboDescription;
}(ToolActionDescription));
export { ToolActionComboDescription };
