// igniteui-react-core/esm5/lib/ToolActionEventDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolActionEventDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolActionEventDetailDescription, _super);
    function ToolActionEventDetailDescription() {
        var _this = _super.call(this) || this;
        _this.p = null;
        _this.q = null;
        _this.i = false;
        _this.h = false;
        _this.n = 0;
        _this.l = new Date();
        _this.s = null;
        _this.r = null;
        return _this;
    }
    ToolActionEventDetailDescription.prototype.get_type = function () {
        return "ToolActionEventDetail";
    };
    Object.defineProperty(ToolActionEventDetailDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionEventDetailDescription.prototype, "actionId", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("ActionId");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionEventDetailDescription.prototype, "actionType", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("ActionType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionEventDetailDescription.prototype, "isModified", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("IsModified");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionEventDetailDescription.prototype, "boolValue", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("BoolValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionEventDetailDescription.prototype, "numberValue", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("NumberValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionEventDetailDescription.prototype, "dateTimeValue", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("DateTimeValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionEventDetailDescription.prototype, "untypedValueRef", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("UntypedValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionEventDetailDescription.prototype, "stringValue", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("StringValue");
        },
        enumerable: false,
        configurable: true
    });
    ToolActionEventDetailDescription.$t = markType(ToolActionEventDetailDescription, 'ToolActionEventDetailDescription', Description.$);
    return ToolActionEventDetailDescription;
}(Description));
export { ToolActionEventDetailDescription };
