// igniteui-react-core/esm5/lib/ToolActionCheckboxDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolActionCheckboxDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolActionCheckboxDescription, _super);
    function ToolActionCheckboxDescription() {
        var _this = _super.call(this) || this;
        _this.dp = false;
        _this.dx = null;
        _this.dt = null;
        _this.dv = null;
        _this.dr = null;
        _this.dy = null;
        _this.du = null;
        _this.dw = null;
        _this.ds = null;
        return _this;
    }
    ToolActionCheckboxDescription.prototype.get_type = function () {
        return "ToolActionCheckbox";
    };
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "isChecked", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("IsChecked");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "uncheckedBackgroundColor", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("UncheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "actualUncheckedBackgroundColor", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("ActualUncheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "checkedBackgroundColor", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.g("CheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "actualCheckedBackgroundColor", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("ActualCheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "uncheckedBorderColor", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("UncheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "actualUncheckedBorderColor", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.g("ActualUncheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "checkedBorderColor", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.g("CheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "actualCheckedBorderColor", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("ActualCheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    ToolActionCheckboxDescription.$t = markType(ToolActionCheckboxDescription, 'ToolActionCheckboxDescription', ToolActionDescription.$);
    return ToolActionCheckboxDescription;
}(ToolActionDescription));
export { ToolActionCheckboxDescription };
