// igniteui-react-core/esm5/lib/ToolContextBindingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolContextBindingDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolContextBindingDescription, _super);
    function ToolContextBindingDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.i = null;
        _this.j = null;
        _this.k = null;
        return _this;
    }
    ToolContextBindingDescription.prototype.get_type = function () {
        return "ToolContextBinding";
    };
    Object.defineProperty(ToolContextBindingDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolContextBindingDescription.prototype, "bindingMode", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("BindingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolContextBindingDescription.prototype, "contextKey", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ContextKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolContextBindingDescription.prototype, "propertyName", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("PropertyName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolContextBindingDescription.prototype, "propertyUpdatedRef", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("PropertyUpdatedRef");
        },
        enumerable: false,
        configurable: true
    });
    ToolContextBindingDescription.$t = markType(ToolContextBindingDescription, 'ToolContextBindingDescription', Description.$);
    return ToolContextBindingDescription;
}(Description));
export { ToolContextBindingDescription };
