// igniteui-react-core/esm5/lib/ToolContextValueType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
export var ToolContextValueType = /*@__PURE__*/ (function (ToolContextValueType) {
    ToolContextValueType[ToolContextValueType["Brush"] = 0] = "Brush";
    ToolContextValueType[ToolContextValueType["BrushCollection"] = 1] = "BrushCollection";
    ToolContextValueType[ToolContextValueType["Color"] = 2] = "Color";
    ToolContextValueType[ToolContextValueType["StringValue"] = 3] = "StringValue";
    ToolContextValueType[ToolContextValueType["DoubleValue"] = 4] = "DoubleValue";
    ToolContextValueType[ToolContextValueType["BoolValue"] = 5] = "BoolValue";
    ToolContextValueType[ToolContextValueType["IntValue"] = 6] = "IntValue";
    ToolContextValueType[ToolContextValueType["Data"] = 7] = "Data";
    return ToolContextValueType;
})({});
/**
 * @hidden
 */
export var ToolContextValueType_$type = /*@__PURE__*/ markEnum('ToolContextValueType', 'Brush,0|BrushCollection,1|Color,2|StringValue,3|DoubleValue,4|BoolValue,5|IntValue,6|Data,7');
