// igniteui-react-core/esm5/lib/ToolActionType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Represents type for toolbar action
 */
export var ToolActionType = /*@__PURE__*/ (function (ToolActionType) {
    ToolActionType[ToolActionType["Unknown"] = 0] = "Unknown";
    ToolActionType[ToolActionType["GroupHeader"] = 1] = "GroupHeader";
    ToolActionType[ToolActionType["Checkbox"] = 2] = "Checkbox";
    ToolActionType[ToolActionType["IconButton"] = 3] = "IconButton";
    ToolActionType[ToolActionType["Label"] = 4] = "Label";
    ToolActionType[ToolActionType["NumberInput"] = 5] = "NumberInput";
    ToolActionType[ToolActionType["Radio"] = 6] = "Radio";
    ToolActionType[ToolActionType["Separator"] = 7] = "Separator";
    ToolActionType[ToolActionType["IconMenu"] = 8] = "IconMenu";
    ToolActionType[ToolActionType["Button"] = 9] = "Button";
    ToolActionType[ToolActionType["SubPanel"] = 10] = "SubPanel";
    ToolActionType[ToolActionType["CheckboxList"] = 11] = "CheckboxList";
    ToolActionType[ToolActionType["ButtonPair"] = 12] = "ButtonPair";
    ToolActionType[ToolActionType["Combo"] = 13] = "Combo";
    ToolActionType[ToolActionType["TextInput"] = 14] = "TextInput";
    ToolActionType[ToolActionType["ColorEditor"] = 15] = "ColorEditor";
    return ToolActionType;
})({});
/**
 * @hidden
 */
export var ToolActionType_$type = /*@__PURE__*/ markEnum('ToolActionType', 'Unknown,0|GroupHeader,1|Checkbox,2|IconButton,3|Label,4|NumberInput,5|Radio,6|Separator,7|IconMenu,8|Button,9|SubPanel,10|CheckboxList,11|ButtonPair,12|Combo,13|TextInput,14|ColorEditor,15');
