// igniteui-react-core/esm5/lib/TransactionEvent.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * An enumeration for the different transaction service events.
 */
export var TransactionEvent = /*@__PURE__*/ (function (TransactionEvent) {
    /**
     * An undo action has occurred.
     */
    TransactionEvent[TransactionEvent["Undo"] = 0] = "Undo";
    /**
     * A redo action has occurred.
     */
    TransactionEvent[TransactionEvent["Redo"] = 1] = "Redo";
    /**
     * Transactions have been cleared from the service.
     */
    TransactionEvent[TransactionEvent["Clear"] = 2] = "Clear";
    /**
     * A transaction was added to the service.
     */
    TransactionEvent[TransactionEvent["Add"] = 3] = "Add";
    /**
     * Pending transactions have ended.
     */
    TransactionEvent[TransactionEvent["End"] = 4] = "End";
    /**
     * Transactions have been committed.
     */
    TransactionEvent[TransactionEvent["Commit"] = 5] = "Commit";
    return TransactionEvent;
})({});
/**
 * @hidden
 */
export var TransactionEvent_$type = /*@__PURE__*/ markEnum('TransactionEvent', 'Undo,0|Redo,1|Clear,2|Add,3|End,4|Commit,5');
