// igniteui-react-core/esm5/lib/TreemapDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var TreemapDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TreemapDescription, _super);
    function TreemapDescription() {
        var _this = _super.call(this) || this;
        _this.da = null;
        _this.c8 = null;
        _this.b1 = null;
        _this.c7 = null;
        _this.cs = null;
        _this.b2 = null;
        _this.cp = null;
        _this.b4 = null;
        _this.cl = null;
        _this.cm = null;
        _this.y = 0;
        _this.cn = null;
        _this.h = null;
        _this.i = null;
        _this.b5 = null;
        _this.l = false;
        _this.bt = 0;
        _this.q = 0;
        _this.p = 0;
        _this.cu = null;
        _this.cr = null;
        _this.c6 = null;
        _this.cg = null;
        _this.ca = null;
        _this.c5 = null;
        _this.cj = null;
        _this.cb = null;
        _this.c9 = null;
        _this.b3 = null;
        _this.ci = null;
        _this.ch = null;
        _this.b7 = null;
        _this.r = 0;
        _this.b8 = null;
        _this.b9 = null;
        _this.cd = null;
        _this.s = 0;
        _this.ce = null;
        _this.cf = null;
        _this.cc = null;
        _this.v = 0;
        _this.x = 0;
        _this.w = 0;
        _this.u = 0;
        _this.ai = 0;
        _this.ak = 0;
        _this.aj = 0;
        _this.ah = 0;
        _this.ac = 0;
        _this.ae = 0;
        _this.ad = 0;
        _this.ab = 0;
        _this.an = 0;
        _this.ar = 0;
        _this.ap = 0;
        _this.al = 0;
        _this.ao = 0;
        _this.as = 0;
        _this.aq = 0;
        _this.am = 0;
        _this.cq = null;
        _this.ct = null;
        _this.t = 0;
        _this.c4 = null;
        _this.au = 0;
        _this.ag = 0;
        _this.cv = null;
        _this.cw = null;
        _this.af = 0;
        _this.bv = 0;
        _this.at = 0;
        _this.o = 0;
        _this.aa = 0;
        _this.n = 0;
        _this.bu = 0;
        _this.co = null;
        _this.bz = null;
        _this.z = 0;
        _this.b6 = null;
        _this.c3 = null;
        _this.c2 = null;
        _this.cz = null;
        _this.cx = null;
        _this.cy = null;
        _this.c0 = null;
        _this.c1 = null;
        _this.db = null;
        _this.ck = null;
        _this.b0 = null;
        return _this;
    }
    TreemapDescription.prototype.get_type = function () {
        return "Treemap";
    };
    Object.defineProperty(TreemapDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.da;
        },
        set: function (a) {
            this.da = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "rootTitle", {
        get: function () {
            return this.c8;
        },
        set: function (a) {
            this.c8 = a;
            this.g("RootTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "breadcrumbSequence", {
        get: function () {
            return this.b1;
        },
        set: function (a) {
            this.b1 = a;
            this.g("BreadcrumbSequence");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "parentIdMemberPath", {
        get: function () {
            return this.c7;
        },
        set: function (a) {
            this.c7 = a;
            this.g("ParentIdMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "labelMemberPath", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("LabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "customValueMemberPath", {
        get: function () {
            return this.b2;
        },
        set: function (a) {
            this.b2 = a;
            this.g("CustomValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "idMemberPath", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("IdMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.b4;
        },
        set: function (a) {
            this.b4 = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "highlightedDataSourceRef", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            this.cl = a;
            this.g("HighlightedDataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "highlightedValueMemberPath", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.g("HighlightedValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "highlightedValueOpacity", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("HighlightedValueOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "highlightedValuesDisplayMode", {
        get: function () {
            return this.cn;
        },
        set: function (a) {
            this.cn = a;
            this.g("HighlightedValuesDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "fillBrushes", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("FillBrushes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "styleMappings", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("StyleMappings");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "fillScaleMode", {
        get: function () {
            return this.b5;
        },
        set: function (a) {
            this.b5 = a;
            this.g("FillScaleMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "isFillScaleLogarithmic", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IsFillScaleLogarithmic");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "fillScaleLogarithmBase", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("FillScaleLogarithmBase");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "fillScaleMinimumValue", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("FillScaleMinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "fillScaleMaximumValue", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("FillScaleMaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "labelVerticalFitMode", {
        get: function () {
            return this.cu;
        },
        set: function (a) {
            this.cu = a;
            this.g("LabelVerticalFitMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "labelHorizontalFitMode", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("LabelHorizontalFitMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "overlayHeaderHoverBackground", {
        get: function () {
            return this.c6;
        },
        set: function (a) {
            this.c6 = a;
            this.g("OverlayHeaderHoverBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerHoverBackground", {
        get: function () {
            return this.cg;
        },
        set: function (a) {
            this.cg = a;
            this.g("HeaderHoverBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerBackground", {
        get: function () {
            return this.ca;
        },
        set: function (a) {
            this.ca = a;
            this.g("HeaderBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "overlayHeaderBackground", {
        get: function () {
            return this.c5;
        },
        set: function (a) {
            this.c5 = a;
            this.g("OverlayHeaderBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerTextColor", {
        get: function () {
            return this.cj;
        },
        set: function (a) {
            this.cj = a;
            this.g("HeaderTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerDarkTextColor", {
        get: function () {
            return this.cb;
        },
        set: function (a) {
            this.cb = a;
            this.g("HeaderDarkTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "textColor", {
        get: function () {
            return this.c9;
        },
        set: function (a) {
            this.c9 = a;
            this.g("TextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "darkTextColor", {
        get: function () {
            return this.b3;
        },
        set: function (a) {
            this.b3 = a;
            this.g("DarkTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerHoverTextColor", {
        get: function () {
            return this.ci;
        },
        set: function (a) {
            this.ci = a;
            this.g("HeaderHoverTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerHoverDarkTextColor", {
        get: function () {
            return this.ch;
        },
        set: function (a) {
            this.ch = a;
            this.g("HeaderHoverDarkTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "fontFamily", {
        get: function () {
            return this.b7;
        },
        set: function (a) {
            this.b7 = a;
            this.g("FontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "fontSize", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("FontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "fontStyle", {
        get: function () {
            return this.b8;
        },
        set: function (a) {
            this.b8 = a;
            this.g("FontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "fontWeight", {
        get: function () {
            return this.b9;
        },
        set: function (a) {
            this.b9 = a;
            this.g("FontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerFontFamily", {
        get: function () {
            return this.cd;
        },
        set: function (a) {
            this.cd = a;
            this.g("HeaderFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerFontSize", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("HeaderFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerFontStyle", {
        get: function () {
            return this.ce;
        },
        set: function (a) {
            this.ce = a;
            this.g("HeaderFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerFontWeight", {
        get: function () {
            return this.cf;
        },
        set: function (a) {
            this.cf = a;
            this.g("HeaderFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerDisplayMode", {
        get: function () {
            return this.cc;
        },
        set: function (a) {
            this.cc = a;
            this.g("HeaderDisplayMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerLabelLeftMargin", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("HeaderLabelLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerLabelTopMargin", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("HeaderLabelTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerLabelRightMargin", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("HeaderLabelRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerLabelBottomMargin", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("HeaderLabelBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "overlayHeaderLabelLeftMargin", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("OverlayHeaderLabelLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "overlayHeaderLabelTopMargin", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("OverlayHeaderLabelTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "overlayHeaderLabelRightMargin", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("OverlayHeaderLabelRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "overlayHeaderLabelBottomMargin", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("OverlayHeaderLabelBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "labelLeftMargin", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("LabelLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "labelTopMargin", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("LabelTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "labelRightMargin", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("LabelRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "labelBottomMargin", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("LabelBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "parentNodeLeftMargin", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("ParentNodeLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "parentNodeTopMargin", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("ParentNodeTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "parentNodeRightMargin", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("ParentNodeRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "parentNodeBottomMargin", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("ParentNodeBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "parentNodeLeftPadding", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("ParentNodeLeftPadding");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "parentNodeTopPadding", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("ParentNodeTopPadding");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "parentNodeRightPadding", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("ParentNodeRightPadding");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "parentNodeBottomPadding", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("ParentNodeBottomPadding");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "labelHorizontalAlignment", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("LabelHorizontalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "labelVerticalAlignment", {
        get: function () {
            return this.ct;
        },
        set: function (a) {
            this.ct = a;
            this.g("LabelVerticalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "headerHeight", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("HeaderHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "outline", {
        get: function () {
            return this.c4;
        },
        set: function (a) {
            this.c4 = a;
            this.g("Outline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "strokeThickness", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("StrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "nodeOpacity", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("NodeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "layoutOrientation", {
        get: function () {
            return this.cv;
        },
        set: function (a) {
            this.cv = a;
            this.g("LayoutOrientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "layoutType", {
        get: function () {
            return this.cw;
        },
        set: function (a) {
            this.cw = a;
            this.g("LayoutType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "minimumDisplaySize", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("MinimumDisplaySize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "transitionDuration", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("TransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("PixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "actualPixelScalingRatio", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("ActualPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "interactionPixelScalingRatio", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("InteractionPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "actualInteractionPixelScalingRatio", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ActualInteractionPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "highlightingTransitionDuration", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("HighlightingTransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "highlightingMode", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.g("HighlightingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "actualHighlightingMode", {
        get: function () {
            return this.bz;
        },
        set: function (a) {
            this.bz = a;
            this.g("ActualHighlightingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "highlightingFadeOpacity", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("HighlightingFadeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "focusItemRef", {
        get: function () {
            return this.b6;
        },
        set: function (a) {
            this.b6 = a;
            this.g("FocusItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "nodeStylingRef", {
        get: function () {
            return this.c3;
        },
        set: function (a) {
            this.c3 = a;
            this.g("NodeStylingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "nodeRenderStylingRef", {
        get: function () {
            return this.c2;
        },
        set: function (a) {
            this.c2 = a;
            this.g("NodeRenderStylingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "nodePointerOverRef", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("NodePointerOverRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "nodePointerEnterRef", {
        get: function () {
            return this.cx;
        },
        set: function (a) {
            this.cx = a;
            this.g("NodePointerEnterRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "nodePointerLeaveRef", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("NodePointerLeaveRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "nodePointerPressedRef", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("NodePointerPressedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "nodePointerReleasedRef", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.g("NodePointerReleasedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "width", {
        get: function () {
            return this.db;
        },
        set: function (a) {
            this.db = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "height", {
        get: function () {
            return this.ck;
        },
        set: function (a) {
            this.ck = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapDescription.prototype, "background", {
        get: function () {
            return this.b0;
        },
        set: function (a) {
            this.b0 = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    TreemapDescription.$t = markType(TreemapDescription, 'TreemapDescription', Description.$);
    return TreemapDescription;
}(Description));
export { TreemapDescription };
