// igniteui-react-core/esm5/lib/TreemapNodePointerEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var TreemapNodePointerEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TreemapNodePointerEventArgsDescription, _super);
    function TreemapNodePointerEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.s = 0;
        _this.r = 0;
        _this.x = null;
        _this.y = null;
        _this.aa = null;
        _this.h = null;
        _this.l = false;
        _this.j = false;
        _this.k = false;
        _this.z = null;
        _this.q = 0;
        _this.ab = null;
        _this.p = 0;
        return _this;
    }
    TreemapNodePointerEventArgsDescription.prototype.get_type = function () {
        return "TreemapNodePointerEventArgs";
    };
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "value", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "sum", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Sum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "customValueRef", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("CustomValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("ItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "parentItemRef", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("ParentItemRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "position", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Position");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "isRightButton", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IsRightButton");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "isHandled", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("IsHandled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "isOverHeader", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("IsOverHeader");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "label", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("Label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "parentValue", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("ParentValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "parentLabel", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("ParentLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreemapNodePointerEventArgsDescription.prototype, "parentSum", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("ParentSum");
        },
        enumerable: false,
        configurable: true
    });
    TreemapNodePointerEventArgsDescription.$t = markType(TreemapNodePointerEventArgsDescription, 'TreemapNodePointerEventArgsDescription', Description.$);
    return TreemapNodePointerEventArgsDescription;
}(Description));
export { TreemapNodePointerEventArgsDescription };
