// igniteui-react-core/esm5/lib/TypeDescriptionWellKnownType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * @hidden
 */
export var TypeDescriptionWellKnownType = /*@__PURE__*/ (function (TypeDescriptionWellKnownType) {
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Void"] = 0] = "Void";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Number"] = 1] = "Number";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["string1"] = 2] = "string1";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Date"] = 3] = "Date";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Brush"] = 4] = "Brush";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Color"] = 5] = "Color";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["BrushCollection"] = 6] = "BrushCollection";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["boolean1"] = 7] = "boolean1";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["ExportedType"] = 8] = "ExportedType";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Collection"] = 9] = "Collection";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Array"] = 10] = "Array";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Point"] = 11] = "Point";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Size"] = 12] = "Size";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["IList"] = 13] = "IList";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Rect"] = 14] = "Rect";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["DataTemplate"] = 15] = "DataTemplate";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["ColorCollection"] = 16] = "ColorCollection";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["Unknown"] = 17] = "Unknown";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["MethodRef"] = 18] = "MethodRef";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["EventRef"] = 19] = "EventRef";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["DataRef"] = 20] = "DataRef";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["TimeSpan"] = 21] = "TimeSpan";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["TemplateRef"] = 22] = "TemplateRef";
    TypeDescriptionWellKnownType[TypeDescriptionWellKnownType["DoubleCollection"] = 23] = "DoubleCollection";
    return TypeDescriptionWellKnownType;
})({});
/**
 * @hidden
 */
export var TypeDescriptionWellKnownType_$type = /*@__PURE__*/ markEnum('TypeDescriptionWellKnownType', 'Void,0|Number,1|String:string1,2|Date,3|Brush,4|Color,5|BrushCollection,6|Boolean:boolean1,7|ExportedType,8|Collection,9|Array,10|Point,11|Size,12|IList,13|Rect,14|DataTemplate,15|ColorCollection,16|Unknown,17|MethodRef,18|EventRef,19|DataRef,20|TimeSpan,21|TemplateRef,22|DoubleCollection,23');
