// igniteui-react-core/esm5/lib/ValueLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ValueLayerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ValueLayerDescription, _super);
    function ValueLayerDescription() {
        var _this = _super.call(this) || this;
        _this.hc = null;
        _this.g7 = null;
        _this.hd = null;
        _this.ha = null;
        _this.hb = null;
        _this.g8 = null;
        _this.g9 = null;
        _this.gb = false;
        _this.f9 = false;
        _this.hf = null;
        _this.hj = null;
        _this.hh = null;
        _this.he = null;
        _this.gf = 0;
        _this.g3 = 0;
        _this.hg = null;
        _this.gh = 0;
        _this.gj = 0;
        _this.gi = 0;
        _this.gg = 0;
        _this.gn = 0;
        _this.gp = 0;
        _this.go = 0;
        _this.gm = 0;
        _this.gk = 0;
        _this.hl = null;
        _this.hi = null;
        _this.gl = 0;
        _this.g4 = 0;
        _this.hk = null;
        _this.gq = 0;
        _this.ga = false;
        return _this;
    }
    ValueLayerDescription.prototype.get_type = function () {
        return "ValueLayer";
    };
    Object.defineProperty(ValueLayerDescription.prototype, "valueMode", {
        get: function () {
            return this.hc;
        },
        set: function (a) {
            this.hc = a;
            this.g("ValueMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "horizontalLineStroke", {
        get: function () {
            return this.g7;
        },
        set: function (a) {
            this.g7 = a;
            this.g("HorizontalLineStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "verticalLineStroke", {
        get: function () {
            return this.hd;
        },
        set: function (a) {
            this.hd = a;
            this.g("VerticalLineStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "targetSeriesName", {
        get: function () {
            return this.ha;
        },
        set: function (a) {
            this.ha = a;
            this.g("TargetSeriesName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "targetSeriesRef", {
        get: function () {
            return this.hb;
        },
        set: function (a) {
            this.hb = a;
            this.g("TargetSeriesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "targetAxisName", {
        get: function () {
            return this.g8;
        },
        set: function (a) {
            this.g8 = a;
            this.g("TargetAxisName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "targetAxisRef", {
        get: function () {
            return this.g9;
        },
        set: function (a) {
            this.g9 = a;
            this.g("TargetAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "useInterpolation", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("UseInterpolation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "isAxisAnnotationEnabled", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("IsAxisAnnotationEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationFormatLabelRef", {
        get: function () {
            return this.hf;
        },
        set: function (a) {
            this.hf = a;
            this.g("XAxisAnnotationFormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationFormatLabelRef", {
        get: function () {
            return this.hj;
        },
        set: function (a) {
            this.hj = a;
            this.g("YAxisAnnotationFormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationTextColor", {
        get: function () {
            return this.hh;
        },
        set: function (a) {
            this.hh = a;
            this.g("XAxisAnnotationTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationBackground", {
        get: function () {
            return this.he;
        },
        set: function (a) {
            this.he = a;
            this.g("XAxisAnnotationBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationBackgroundCornerRadius", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("XAxisAnnotationBackgroundCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationInterpolatedValuePrecision", {
        get: function () {
            return this.g3;
        },
        set: function (a) {
            this.g3 = a;
            this.g("XAxisAnnotationInterpolatedValuePrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationOutline", {
        get: function () {
            return this.hg;
        },
        set: function (a) {
            this.hg = a;
            this.g("XAxisAnnotationOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationPaddingLeft", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("XAxisAnnotationPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationPaddingTop", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("XAxisAnnotationPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationPaddingRight", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("XAxisAnnotationPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationPaddingBottom", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("XAxisAnnotationPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationPaddingLeft", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("YAxisAnnotationPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationPaddingTop", {
        get: function () {
            return this.gp;
        },
        set: function (a) {
            this.gp = a;
            this.g("YAxisAnnotationPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationPaddingRight", {
        get: function () {
            return this.go;
        },
        set: function (a) {
            this.go = a;
            this.g("YAxisAnnotationPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationPaddingBottom", {
        get: function () {
            return this.gm;
        },
        set: function (a) {
            this.gm = a;
            this.g("YAxisAnnotationPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "xAxisAnnotationStrokeThickness", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("XAxisAnnotationStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationTextColor", {
        get: function () {
            return this.hl;
        },
        set: function (a) {
            this.hl = a;
            this.g("YAxisAnnotationTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationBackground", {
        get: function () {
            return this.hi;
        },
        set: function (a) {
            this.hi = a;
            this.g("YAxisAnnotationBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationBackgroundCornerRadius", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("YAxisAnnotationBackgroundCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationInterpolatedValuePrecision", {
        get: function () {
            return this.g4;
        },
        set: function (a) {
            this.g4 = a;
            this.g("YAxisAnnotationInterpolatedValuePrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationOutline", {
        get: function () {
            return this.hk;
        },
        set: function (a) {
            this.hk = a;
            this.g("YAxisAnnotationOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "yAxisAnnotationStrokeThickness", {
        get: function () {
            return this.gq;
        },
        set: function (a) {
            this.gq = a;
            this.g("YAxisAnnotationStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayerDescription.prototype, "skipUnknownValues", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("SkipUnknownValues");
        },
        enumerable: false,
        configurable: true
    });
    ValueLayerDescription.$t = markType(ValueLayerDescription, 'ValueLayerDescription', AnnotationLayerDescription.$);
    return ValueLayerDescription;
}(AnnotationLayerDescription));
export { ValueLayerDescription };
