// igniteui-react-core/esm5/lib/VirtualDataSourceDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseDataSourceDescription } from "./BaseDataSourceDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var VirtualDataSourceDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(VirtualDataSourceDescription, _super);
    function VirtualDataSourceDescription() {
        var _this = _super.call(this) || this;
        _this.a3 = false;
        _this.a4 = false;
        _this.ba = 0;
        _this.a9 = 0;
        _this.a8 = 0;
        _this.be = null;
        _this.a2 = false;
        return _this;
    }
    VirtualDataSourceDescription.prototype.get_type = function () {
        return "VirtualDataSource";
    };
    Object.defineProperty(VirtualDataSourceDescription.prototype, "isSectionCollapsable", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("IsSectionCollapsable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(VirtualDataSourceDescription.prototype, "isSectionExpandedDefault", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("IsSectionExpandedDefault");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(VirtualDataSourceDescription.prototype, "pageSizeRequested", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("PageSizeRequested");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(VirtualDataSourceDescription.prototype, "maxCachedPages", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("MaxCachedPages");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(VirtualDataSourceDescription.prototype, "actualPageSize", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("ActualPageSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(VirtualDataSourceDescription.prototype, "concurrencyTag", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.g("ConcurrencyTag");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(VirtualDataSourceDescription.prototype, "isBatchingEnabled", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("IsBatchingEnabled");
        },
        enumerable: false,
        configurable: true
    });
    VirtualDataSourceDescription.$t = markType(VirtualDataSourceDescription, 'VirtualDataSourceDescription', BaseDataSourceDescription.$);
    return VirtualDataSourceDescription;
}(BaseDataSourceDescription));
export { VirtualDataSourceDescription };
