// igniteui-react-core/esm5/lib/WaterfallSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { HorizontalAnchoredCategorySeriesDescription } from "./HorizontalAnchoredCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WaterfallSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WaterfallSeriesDescription, _super);
    function WaterfallSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.g3 = null;
        _this.gz = 0;
        _this.g0 = 0;
        _this.g4 = null;
        return _this;
    }
    WaterfallSeriesDescription.prototype.get_type = function () {
        return "WaterfallSeries";
    };
    Object.defineProperty(WaterfallSeriesDescription.prototype, "negativeBrush", {
        get: function () {
            return this.g3;
        },
        set: function (a) {
            this.g3 = a;
            this.g("NegativeBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WaterfallSeriesDescription.prototype, "radiusX", {
        get: function () {
            return this.gz;
        },
        set: function (a) {
            this.gz = a;
            this.g("RadiusX");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WaterfallSeriesDescription.prototype, "radiusY", {
        get: function () {
            return this.g0;
        },
        set: function (a) {
            this.g0 = a;
            this.g("RadiusY");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WaterfallSeriesDescription.prototype, "negativeOutline", {
        get: function () {
            return this.g4;
        },
        set: function (a) {
            this.g4 = a;
            this.g("NegativeOutline");
        },
        enumerable: false,
        configurable: true
    });
    WaterfallSeriesDescription.$t = markType(WaterfallSeriesDescription, 'WaterfallSeriesDescription', HorizontalAnchoredCategorySeriesDescription.$);
    return WaterfallSeriesDescription;
}(HorizontalAnchoredCategorySeriesDescription));
export { WaterfallSeriesDescription };
