// igniteui-react-core/esm5/lib/WebAvatarDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebAvatarDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebAvatarDescription, _super);
    function WebAvatarDescription() {
        var _this = _super.call(this) || this;
        _this.k = null;
        _this.h = null;
        _this.i = null;
        _this.j = null;
        return _this;
    }
    WebAvatarDescription.prototype.get_type = function () {
        return "WebAvatar";
    };
    Object.defineProperty(WebAvatarDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebAvatarDescription.prototype, "src", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Src");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebAvatarDescription.prototype, "alt", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Alt");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebAvatarDescription.prototype, "initials", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Initials");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebAvatarDescription.prototype, "shape", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Shape");
        },
        enumerable: false,
        configurable: true
    });
    WebAvatarDescription.$t = markType(WebAvatarDescription, 'WebAvatarDescription', Description.$);
    return WebAvatarDescription;
}(Description));
export { WebAvatarDescription };
