// igniteui-react-core/esm5/lib/WebActiveNodeChangeEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebActiveNodeChangeEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebActiveNodeChangeEventArgsDetailDescription, _super);
    function WebActiveNodeChangeEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.p = 0;
        _this.n = 0;
        _this.o = 0;
        _this.u = null;
        return _this;
    }
    WebActiveNodeChangeEventArgsDetailDescription.prototype.get_type = function () {
        return "WebActiveNodeChangeEventArgsDetail";
    };
    Object.defineProperty(WebActiveNodeChangeEventArgsDetailDescription.prototype, "row", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Row");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebActiveNodeChangeEventArgsDetailDescription.prototype, "column", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Column");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebActiveNodeChangeEventArgsDetailDescription.prototype, "level", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Level");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebActiveNodeChangeEventArgsDetailDescription.prototype, "tag", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Tag");
        },
        enumerable: false,
        configurable: true
    });
    WebActiveNodeChangeEventArgsDetailDescription.$t = markType(WebActiveNodeChangeEventArgsDetailDescription, 'WebActiveNodeChangeEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebActiveNodeChangeEventArgsDetailDescription.__marshalByValue1 = true;
    WebActiveNodeChangeEventArgsDetailDescription.__marshalByValueAlias1 = "ActiveNodeChangeEventArgsDetail";
    return WebActiveNodeChangeEventArgsDetailDescription;
}(WebBaseEventArgsDetailDescription));
export { WebActiveNodeChangeEventArgsDetailDescription };
