// igniteui-react-core/esm5/lib/WebBaseAlertLikeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebBaseAlertLikeDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebBaseAlertLikeDescription, _super);
    function WebBaseAlertLikeDescription() {
        var _this = _super.call(this) || this;
        _this.i = false;
        _this.l = 0;
        _this.h = false;
        _this.n = null;
        return _this;
    }
    WebBaseAlertLikeDescription.prototype.get_type = function () {
        return "WebBaseAlertLike";
    };
    Object.defineProperty(WebBaseAlertLikeDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseAlertLikeDescription.prototype, "open", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Open");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseAlertLikeDescription.prototype, "displayTime", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("DisplayTime");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseAlertLikeDescription.prototype, "keepOpen", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("KeepOpen");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseAlertLikeDescription.prototype, "position", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Position");
        },
        enumerable: false,
        configurable: true
    });
    WebBaseAlertLikeDescription.$t = markType(WebBaseAlertLikeDescription, 'WebBaseAlertLikeDescription', Description.$);
    return WebBaseAlertLikeDescription;
}(Description));
export { WebBaseAlertLikeDescription };
