// igniteui-react-core/esm5/lib/WebBaseOptionLikeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebBaseOptionLikeDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebBaseOptionLikeDescription, _super);
    function WebBaseOptionLikeDescription() {
        var _this = _super.call(this) || this;
        _this.h = false;
        _this.i = false;
        _this.j = false;
        _this.n = null;
        return _this;
    }
    WebBaseOptionLikeDescription.prototype.get_type = function () {
        return "WebBaseOptionLike";
    };
    Object.defineProperty(WebBaseOptionLikeDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseOptionLikeDescription.prototype, "active", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Active");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseOptionLikeDescription.prototype, "disabled", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseOptionLikeDescription.prototype, "selected", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Selected");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseOptionLikeDescription.prototype, "value", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    WebBaseOptionLikeDescription.$t = markType(WebBaseOptionLikeDescription, 'WebBaseOptionLikeDescription', Description.$);
    return WebBaseOptionLikeDescription;
}(Description));
export { WebBaseOptionLikeDescription };
