// igniteui-react-core/esm5/lib/WebBaseSearchInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebBaseSearchInfoDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebBaseSearchInfoDescription, _super);
    function WebBaseSearchInfoDescription() {
        var _this = _super.call(this) || this;
        _this.o = null;
        _this.h = false;
        _this.i = false;
        _this.l = 0;
        _this.n = null;
        return _this;
    }
    WebBaseSearchInfoDescription.prototype.get_type = function () {
        return "WebBaseSearchInfo";
    };
    Object.defineProperty(WebBaseSearchInfoDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseSearchInfoDescription.prototype, "searchText", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("SearchText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseSearchInfoDescription.prototype, "caseSensitive", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("CaseSensitive");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseSearchInfoDescription.prototype, "exactMatch", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ExactMatch");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseSearchInfoDescription.prototype, "matchCount", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("MatchCount");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseSearchInfoDescription.prototype, "content", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Content");
        },
        enumerable: false,
        configurable: true
    });
    WebBaseSearchInfoDescription.$t = markType(WebBaseSearchInfoDescription, 'WebBaseSearchInfoDescription', Description.$);
    return WebBaseSearchInfoDescription;
}(Description));
export { WebBaseSearchInfoDescription };
