// igniteui-react-core/esm5/lib/WebBaseComboBoxLikeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebBaseComboBoxLikeDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebBaseComboBoxLikeDescription, _super);
    function WebBaseComboBoxLikeDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.i = false;
        _this.h = false;
        _this.j = false;
        return _this;
    }
    WebBaseComboBoxLikeDescription.prototype.get_type = function () {
        return "WebBaseComboBoxLike";
    };
    Object.defineProperty(WebBaseComboBoxLikeDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseComboBoxLikeDescription.prototype, "emitEvent", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("EmitEvent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseComboBoxLikeDescription.prototype, "keepOpenOnSelect", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("KeepOpenOnSelect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseComboBoxLikeDescription.prototype, "keepOpenOnOutsideClick", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("KeepOpenOnOutsideClick");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebBaseComboBoxLikeDescription.prototype, "open", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Open");
        },
        enumerable: false,
        configurable: true
    });
    WebBaseComboBoxLikeDescription.$t = markType(WebBaseComboBoxLikeDescription, 'WebBaseComboBoxLikeDescription', Description.$);
    return WebBaseComboBoxLikeDescription;
}(Description));
export { WebBaseComboBoxLikeDescription };
