// igniteui-react-core/esm5/lib/WebButtonGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebButtonGroupDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebButtonGroupDescription, _super);
    function WebButtonGroupDescription() {
        var _this = _super.call(this) || this;
        _this.j = false;
        _this.l = null;
        _this.o = null;
        _this.h = null;
        _this.n = null;
        _this.m = null;
        return _this;
    }
    WebButtonGroupDescription.prototype.get_type = function () {
        return "WebButtonGroup";
    };
    Object.defineProperty(WebButtonGroupDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonGroupDescription.prototype, "disabled", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonGroupDescription.prototype, "alignment", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Alignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonGroupDescription.prototype, "selection", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Selection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonGroupDescription.prototype, "selectedItems", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("SelectedItems");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonGroupDescription.prototype, "selectRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("SelectRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebButtonGroupDescription.prototype, "deselectRef", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("DeselectRef");
        },
        enumerable: false,
        configurable: true
    });
    WebButtonGroupDescription.$t = markType(WebButtonGroupDescription, 'WebButtonGroupDescription', Description.$);
    return WebButtonGroupDescription;
}(Description));
export { WebButtonGroupDescription };
