// igniteui-react-core/esm5/lib/WebColumnGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebColumnDescription } from "./WebColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebColumnGroupDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebColumnGroupDescription, _super);
    function WebColumnGroupDescription() {
        var _this = _super.call(this) || this;
        _this.cm = null;
        _this.cp = false;
        _this.cq = false;
        _this.cu = null;
        return _this;
    }
    WebColumnGroupDescription.prototype.get_type = function () {
        return "WebColumnGroup";
    };
    Object.defineProperty(WebColumnGroupDescription.prototype, "children", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.g("Children");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnGroupDescription.prototype, "collapsible", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("Collapsible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnGroupDescription.prototype, "expanded", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("Expanded");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnGroupDescription.prototype, "collapsibleIndicatorTemplateRef", {
        get: function () {
            return this.cu;
        },
        set: function (a) {
            this.cu = a;
            this.g("CollapsibleIndicatorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    WebColumnGroupDescription.$t = markType(WebColumnGroupDescription, 'WebColumnGroupDescription', WebColumnDescription.$);
    WebColumnGroupDescription.__canMarshalByValue1 = true;
    WebColumnGroupDescription.__marshalByValueAlias1 = "ColumnGroup";
    return WebColumnGroupDescription;
}(WebColumnDescription));
export { WebColumnGroupDescription };
