// igniteui-react-core/esm5/lib/WebChildGridRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebChildGridRowDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebChildGridRowDescription, _super);
    function WebChildGridRowDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.n = null;
        _this.l = 0;
        _this.j = false;
        return _this;
    }
    WebChildGridRowDescription.prototype.get_type = function () {
        return "WebChildGridRow";
    };
    Object.defineProperty(WebChildGridRowDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebChildGridRowDescription.prototype, "layout", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Layout");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebChildGridRowDescription.prototype, "data", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Data");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebChildGridRowDescription.prototype, "index", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Index");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebChildGridRowDescription.prototype, "expanded", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Expanded");
        },
        enumerable: false,
        configurable: true
    });
    WebChildGridRowDescription.$t = markType(WebChildGridRowDescription, 'WebChildGridRowDescription', Description.$);
    return WebChildGridRowDescription;
}(Description));
export { WebChildGridRowDescription };
