// igniteui-react-core/esm5/lib/WebComboDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebComboDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebComboDescription, _super);
    function WebComboDescription() {
        var _this = _super.call(this) || this;
        _this.al = null;
        _this.s = false;
        _this.u = false;
        _this.l = false;
        _this.m = false;
        _this.as = null;
        _this.aw = null;
        _this.ax = null;
        _this.r = false;
        _this.ay = null;
        _this.am = null;
        _this.ap = null;
        _this.aq = null;
        _this.j = null;
        _this.n = false;
        _this.p = false;
        _this.h = null;
        _this.o = false;
        _this.t = false;
        _this.af = null;
        _this.at = null;
        _this.q = false;
        _this.ar = null;
        _this.ao = null;
        _this.ai = null;
        _this.an = null;
        _this.ah = null;
        _this.av = null;
        _this.au = null;
        _this.ak = null;
        _this.aj = null;
        return _this;
    }
    WebComboDescription.prototype.get_type = function () {
        return "WebCombo";
    };
    Object.defineProperty(WebComboDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "dataRef", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("DataRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "outlined", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Outlined");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "singleSelect", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("SingleSelect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "autofocus", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Autofocus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "autofocusList", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("AutofocusList");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "label", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("Label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "placeholder", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("Placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "placeholderSearch", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("PlaceholderSearch");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "open", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Open");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "valueKey", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("ValueKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "displayKey", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("DisplayKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "groupKey", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("GroupKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "groupSorting", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("GroupSorting");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "filteringOptions", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("FilteringOptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "caseSensitiveIcon", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("CaseSensitiveIcon");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "disableFiltering", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("DisableFiltering");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "value", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "disabled", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "required", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("Required");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "defaultValue", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("DefaultValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "name", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "invalid", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Invalid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "itemTemplateRef", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("ItemTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "groupHeaderTemplateRef", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("GroupHeaderTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "changeRef", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "focusRef", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("FocusRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "blurRef", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("BlurRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "openingRef", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("OpeningRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "openedRef", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("OpenedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "closingRef", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("ClosingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebComboDescription.prototype, "closedRef", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("ClosedRef");
        },
        enumerable: false,
        configurable: true
    });
    WebComboDescription.$t = markType(WebComboDescription, 'WebComboDescription', Description.$);
    return WebComboDescription;
}(Description));
export { WebComboDescription };
