// igniteui-react-core/esm5/lib/WebCalendarBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebCalendarBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebCalendarBaseDescription, _super);
    function WebCalendarBaseDescription() {
        var _this = _super.call(this) || this;
        _this.o = null;
        _this.l = false;
        _this.p = null;
        _this.n = null;
        _this.i = null;
        _this.h = null;
        return _this;
    }
    WebCalendarBaseDescription.prototype.get_type = function () {
        return "WebCalendarBase";
    };
    Object.defineProperty(WebCalendarBaseDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarBaseDescription.prototype, "selection", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Selection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarBaseDescription.prototype, "showWeekNumbers", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ShowWeekNumbers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarBaseDescription.prototype, "weekStart", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("WeekStart");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarBaseDescription.prototype, "locale", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Locale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarBaseDescription.prototype, "specialDates", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("SpecialDates");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarBaseDescription.prototype, "disabledDates", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("DisabledDates");
        },
        enumerable: false,
        configurable: true
    });
    WebCalendarBaseDescription.$t = markType(WebCalendarBaseDescription, 'WebCalendarBaseDescription', Description.$);
    return WebCalendarBaseDescription;
}(Description));
export { WebCalendarBaseDescription };
