// igniteui-react-core/esm5/lib/WebClipboardOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebClipboardOptionsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebClipboardOptionsDescription, _super);
    function WebClipboardOptionsDescription() {
        var _this = _super.call(this) || this;
        _this.j = false;
        _this.i = false;
        _this.h = false;
        _this.n = null;
        return _this;
    }
    WebClipboardOptionsDescription.prototype.get_type = function () {
        return "WebClipboardOptions";
    };
    Object.defineProperty(WebClipboardOptionsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebClipboardOptionsDescription.prototype, "enabled", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Enabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebClipboardOptionsDescription.prototype, "copyHeaders", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("CopyHeaders");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebClipboardOptionsDescription.prototype, "copyFormatters", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("CopyFormatters");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebClipboardOptionsDescription.prototype, "separator", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Separator");
        },
        enumerable: false,
        configurable: true
    });
    WebClipboardOptionsDescription.$t = markType(WebClipboardOptionsDescription, 'WebClipboardOptionsDescription', Description.$);
    return WebClipboardOptionsDescription;
}(Description));
export { WebClipboardOptionsDescription };
