// igniteui-react-core/esm5/lib/WebCalendarResourceStringsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebCalendarResourceStringsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebCalendarResourceStringsDescription, _super);
    function WebCalendarResourceStringsDescription() {
        var _this = _super.call(this) || this;
        _this.s = null;
        _this.u = null;
        _this.q = null;
        _this.t = null;
        _this.r = null;
        _this.v = null;
        _this.j = null;
        _this.n = null;
        _this.k = null;
        _this.o = null;
        _this.l = null;
        _this.p = null;
        _this.m = null;
        _this.w = null;
        return _this;
    }
    WebCalendarResourceStringsDescription.prototype.get_type = function () {
        return "WebCalendarResourceStrings";
    };
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "selectMonth", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("SelectMonth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "selectYear", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("SelectYear");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "selectDate", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("SelectDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "selectRange", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("SelectRange");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "selectedDate", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("SelectedDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "startDate", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("StartDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "endDate", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("EndDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "previousMonth", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("PreviousMonth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "nextMonth", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("NextMonth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "previousYear", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("PreviousYear");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "nextYear", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("NextYear");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "previousYears", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("PreviousYears");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "nextYears", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("NextYears");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarResourceStringsDescription.prototype, "weekLabel", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("WeekLabel");
        },
        enumerable: false,
        configurable: true
    });
    WebCalendarResourceStringsDescription.$t = markType(WebCalendarResourceStringsDescription, 'WebCalendarResourceStringsDescription', Description.$);
    WebCalendarResourceStringsDescription.__marshalByValue = true;
    WebCalendarResourceStringsDescription.__marshalByValueAlias = "CalendarResourceStrings";
    return WebCalendarResourceStringsDescription;
}(Description));
export { WebCalendarResourceStringsDescription };
