// igniteui-react-core/esm5/lib/WebColumnExportingEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebColumnExportingEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebColumnExportingEventArgsDetailDescription, _super);
    function WebColumnExportingEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.x = null;
        _this.w = null;
        _this.t = 0;
        _this.p = false;
        _this.q = false;
        _this.m = null;
        return _this;
    }
    WebColumnExportingEventArgsDetailDescription.prototype.get_type = function () {
        return "WebColumnExportingEventArgsDetail";
    };
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "header", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("Header");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "field", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("Field");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "columnIndex", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("ColumnIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "cancel", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Cancel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "skipFormatter", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("SkipFormatter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "grid", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Grid");
        },
        enumerable: false,
        configurable: true
    });
    WebColumnExportingEventArgsDetailDescription.$t = markType(WebColumnExportingEventArgsDetailDescription, 'WebColumnExportingEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebColumnExportingEventArgsDetailDescription.__marshalByValue1 = true;
    WebColumnExportingEventArgsDetailDescription.__marshalByValueAlias1 = "ColumnExportingEventArgsDetail";
    return WebColumnExportingEventArgsDetailDescription;
}(WebBaseEventArgsDetailDescription));
export { WebColumnExportingEventArgsDetailDescription };
