// igniteui-react-core/esm5/lib/WebCarouselDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebCarouselDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebCarouselDescription, _super);
    function WebCarouselDescription() {
        var _this = _super.call(this) || this;
        _this.h = false;
        _this.i = false;
        _this.k = false;
        _this.j = false;
        _this.l = false;
        _this.x = null;
        _this.w = null;
        _this.ab = null;
        _this.r = 0;
        _this.s = 0;
        _this.v = null;
        _this.aa = null;
        _this.z = null;
        _this.y = null;
        return _this;
    }
    WebCarouselDescription.prototype.get_type = function () {
        return "WebCarousel";
    };
    Object.defineProperty(WebCarouselDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "disableLoop", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("DisableLoop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "disablePauseOnInteraction", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("DisablePauseOnInteraction");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "hideNavigation", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("HideNavigation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "hideIndicators", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("HideIndicators");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "vertical", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Vertical");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "indicatorsOrientation", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("IndicatorsOrientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "indicatorsLabelFormat", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("IndicatorsLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "slidesLabelFormat", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("SlidesLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "interval", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Interval");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "maximumIndicatorsCount", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("MaximumIndicatorsCount");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "animationType", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("AnimationType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "slideChangedRef", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("SlideChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "playingRef", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("PlayingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCarouselDescription.prototype, "pausedRef", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("PausedRef");
        },
        enumerable: false,
        configurable: true
    });
    WebCarouselDescription.$t = markType(WebCarouselDescription, 'WebCarouselDescription', Description.$);
    return WebCarouselDescription;
}(Description));
export { WebCarouselDescription };
