// igniteui-react-core/esm5/lib/WebCalendarDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebCalendarBaseDescription } from "./WebCalendarBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebCalendarDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebCalendarDescription, _super);
    function WebCalendarDescription() {
        var _this = _super.call(this) || this;
        _this.af = new Date();
        _this.u = null;
        _this.ae = new Date();
        _this.ab = false;
        _this.aa = false;
        _this.am = null;
        _this.an = null;
        _this.ai = 0;
        _this.ak = null;
        _this.w = null;
        _this.y = null;
        _this.al = null;
        return _this;
    }
    WebCalendarDescription.prototype.get_type = function () {
        return "WebCalendar";
    };
    Object.defineProperty(WebCalendarDescription.prototype, "value", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "values", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Values");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "activeDate", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("ActiveDate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "hideOutsideDays", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("HideOutsideDays");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "hideHeader", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("HideHeader");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "headerOrientation", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("HeaderOrientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "orientation", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("Orientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "visibleMonths", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("VisibleMonths");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "activeView", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("ActiveView");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "formatOptions", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("FormatOptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "resourceStrings", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("ResourceStrings");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebCalendarDescription.prototype, "changeRef", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    WebCalendarDescription.$t = markType(WebCalendarDescription, 'WebCalendarDescription', WebCalendarBaseDescription.$);
    return WebCalendarDescription;
}(WebCalendarBaseDescription));
export { WebCalendarDescription };
