// igniteui-react-core/esm5/lib/WebColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebColumnDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebColumnDescription, _super);
    function WebColumnDescription() {
        var _this = _super.call(this) || this;
        _this.bk = null;
        _this.bn = null;
        _this.by = null;
        _this.ae = false;
        _this.ac = false;
        _this.w = false;
        _this.t = false;
        _this.u = false;
        _this.aa = false;
        _this.q = false;
        _this.x = false;
        _this.y = false;
        _this.ad = false;
        _this.r = false;
        _this.s = false;
        _this.bz = null;
        _this.bs = null;
        _this.bo = null;
        _this.a8 = null;
        _this.bp = null;
        _this.a7 = null;
        _this.bg = null;
        _this.bh = null;
        _this.bm = null;
        _this.bw = null;
        _this.v = false;
        _this.af = false;
        _this.ab = false;
        _this.bi = null;
        _this.a0 = 0;
        _this.ay = 0;
        _this.a1 = 0;
        _this.az = 0;
        _this.a6 = null;
        _this.bt = null;
        _this.z = false;
        _this.bv = null;
        _this.l = null;
        _this.n = null;
        _this.bx = null;
        _this.bf = null;
        _this.bq = null;
        _this.br = null;
        _this.bj = null;
        _this.bl = null;
        _this.ag = false;
        _this.j = null;
        _this.h = null;
        _this.a9 = null;
        _this.bu = null;
        return _this;
    }
    WebColumnDescription.prototype.get_type = function () {
        return "WebColumn";
    };
    Object.defineProperty(WebColumnDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "field", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("Field");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "header", {
        get: function () {
            return this.bn;
        },
        set: function (a) {
            this.bn = a;
            this.g("Header");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "title", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            this.by = a;
            this.g("Title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "sortable", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("Sortable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "selectable", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("Selectable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "groupable", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("Groupable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "editable", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("Editable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "filterable", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Filterable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "resizable", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("Resizable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "autosizeHeader", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("AutosizeHeader");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "hasSummary", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("HasSummary");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "hidden", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("Hidden");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "selected", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("Selected");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "disableHiding", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("DisableHiding");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "disablePinning", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("DisablePinning");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "width", {
        get: function () {
            return this.bz;
        },
        set: function (a) {
            this.bz = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "maxWidth", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("MaxWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "headerClasses", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            this.bo = a;
            this.g("HeaderClasses");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "headerStyles", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("HeaderStyles");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "headerGroupClasses", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.g("HeaderGroupClasses");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "headerGroupStyles", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("HeaderGroupStyles");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "cellClassesRef", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.g("CellClassesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "cellStylesRef", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.g("CellStylesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "formatterRef", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            this.bm = a;
            this.g("FormatterRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "summaryFormatterRef", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("SummaryFormatterRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "filteringIgnoreCase", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("FilteringIgnoreCase");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "sortingIgnoreCase", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("SortingIgnoreCase");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "searchable", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("Searchable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "dataType", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("DataType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "rowEnd", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("RowEnd");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "colEnd", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("ColEnd");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "rowStart", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("RowStart");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "colStart", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("ColStart");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "additionalTemplateContext", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("AdditionalTemplateContext");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "minWidth", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("MinWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "pinned", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("Pinned");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "summariesRef", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("SummariesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "filters", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Filters");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "sortStrategy", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("SortStrategy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "summaryTemplateRef", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("SummaryTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "bodyTemplateRef", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("BodyTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "headerTemplateRef", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            this.bq = a;
            this.g("HeaderTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "inlineEditorTemplateRef", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.g("InlineEditorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "errorTemplateRef", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("ErrorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "filterCellTemplateRef", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.g("FilterCellTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "visibleWhenCollapsed", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("VisibleWhenCollapsed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "pipeArgs", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("PipeArgs");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "editorOptions", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("EditorOptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "parentColumn", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("ParentColumn");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnDescription.prototype, "name", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    WebColumnDescription.$t = markType(WebColumnDescription, 'WebColumnDescription', Description.$);
    WebColumnDescription.__canMarshalByValue = true;
    WebColumnDescription.__marshalByValueAlias = "Column";
    return WebColumnDescription;
}(Description));
export { WebColumnDescription };
