// igniteui-react-core/esm5/lib/WebDockManagerResourceStringsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebDockManagerResourceStringsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebDockManagerResourceStringsDescription, _super);
    function WebDockManagerResourceStringsDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.o = null;
        _this.p = null;
        _this.j = null;
        _this.k = null;
        _this.l = null;
        _this.m = null;
        _this.n = null;
        _this.i = null;
        return _this;
    }
    WebDockManagerResourceStringsDescription.prototype.get_type = function () {
        return "WebDockManagerResourceStrings";
    };
    Object.defineProperty(WebDockManagerResourceStringsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerResourceStringsDescription.prototype, "close", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Close");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerResourceStringsDescription.prototype, "pin", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Pin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerResourceStringsDescription.prototype, "unpin", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Unpin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerResourceStringsDescription.prototype, "maximize", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Maximize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerResourceStringsDescription.prototype, "minimize", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Minimize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerResourceStringsDescription.prototype, "moreOptions", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("MoreOptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerResourceStringsDescription.prototype, "moreTabs", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("MoreTabs");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerResourceStringsDescription.prototype, "panes", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Panes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerResourceStringsDescription.prototype, "documents", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Documents");
        },
        enumerable: false,
        configurable: true
    });
    WebDockManagerResourceStringsDescription.$t = markType(WebDockManagerResourceStringsDescription, 'WebDockManagerResourceStringsDescription', Description.$);
    return WebDockManagerResourceStringsDescription;
}(Description));
export { WebDockManagerResourceStringsDescription };
