// igniteui-react-core/esm5/lib/WebDialogDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebDialogDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebDialogDescription, _super);
    function WebDialogDescription() {
        var _this = _super.call(this) || this;
        _this.j = false;
        _this.h = false;
        _this.i = false;
        _this.k = false;
        _this.s = null;
        _this.r = null;
        _this.q = null;
        _this.p = null;
        return _this;
    }
    WebDialogDescription.prototype.get_type = function () {
        return "WebDialog";
    };
    Object.defineProperty(WebDialogDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDialogDescription.prototype, "keepOpenOnEscape", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("KeepOpenOnEscape");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDialogDescription.prototype, "closeOnOutsideClick", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("CloseOnOutsideClick");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDialogDescription.prototype, "hideDefaultAction", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("HideDefaultAction");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDialogDescription.prototype, "open", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Open");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDialogDescription.prototype, "title", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDialogDescription.prototype, "returnValue", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("ReturnValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDialogDescription.prototype, "closingRef", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("ClosingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDialogDescription.prototype, "closedRef", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("ClosedRef");
        },
        enumerable: false,
        configurable: true
    });
    WebDialogDescription.$t = markType(WebDialogDescription, 'WebDialogDescription', Description.$);
    return WebDialogDescription;
}(Description));
export { WebDialogDescription };
