// igniteui-react-core/esm5/lib/WebDockManagerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebDockManagerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebDockManagerDescription, _super);
    function WebDockManagerDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.l = null;
        _this.p = null;
        _this.h = null;
        _this.v = false;
        _this.w = false;
        _this.ah = null;
        _this.m = null;
        _this.r = null;
        _this.t = false;
        _this.x = false;
        _this.u = false;
        _this.ai = null;
        _this.y = false;
        _this.af = null;
        _this.aj = null;
        _this.ag = null;
        return _this;
    }
    WebDockManagerDescription.prototype.get_type = function () {
        return "WebDockManager";
    };
    Object.defineProperty(WebDockManagerDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "layout", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Layout");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "draggedPane", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("DraggedPane");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "dropPosition", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("DropPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "activePane", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ActivePane");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "allowMaximize", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("AllowMaximize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "containedInBoundaries", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("ContainedInBoundaries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "showHeaderIconOnHover", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("ShowHeaderIconOnHover");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "maximizedPane", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("MaximizedPane");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "resourceStrings", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("ResourceStrings");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "allowFloatingPanesResize", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("AllowFloatingPanesResize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "disableKeyboardNavigation", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("DisableKeyboardNavigation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "allowInnerDock", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("AllowInnerDock");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "showPaneHeaders", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("ShowPaneHeaders");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "proximityDock", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("ProximityDock");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "closeBehavior", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("CloseBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "unpinBehavior", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("UnpinBehavior");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebDockManagerDescription.prototype, "layoutChangeRef", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("LayoutChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    WebDockManagerDescription.$t = markType(WebDockManagerDescription, 'WebDockManagerDescription', Description.$);
    return WebDockManagerDescription;
}(Description));
export { WebDockManagerDescription };
