// igniteui-react-core/esm5/lib/WebFilteringOperationDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebFilteringOperationDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebFilteringOperationDescription, _super);
    function WebFilteringOperationDescription() {
        var _this = _super.call(this) || this;
        _this.p = null;
        _this.j = false;
        _this.n = null;
        _this.i = false;
        _this.o = null;
        return _this;
    }
    WebFilteringOperationDescription.prototype.get_type = function () {
        return "WebFilteringOperation";
    };
    Object.defineProperty(WebFilteringOperationDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringOperationDescription.prototype, "name", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringOperationDescription.prototype, "isUnary", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("IsUnary");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringOperationDescription.prototype, "iconName", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("IconName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringOperationDescription.prototype, "hidden", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Hidden");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringOperationDescription.prototype, "logicRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("LogicRef");
        },
        enumerable: false,
        configurable: true
    });
    WebFilteringOperationDescription.$t = markType(WebFilteringOperationDescription, 'WebFilteringOperationDescription', Description.$);
    WebFilteringOperationDescription.__marshalByValue = true;
    WebFilteringOperationDescription.__marshalByValueAlias = "FilteringOperation";
    return WebFilteringOperationDescription;
}(Description));
export { WebFilteringOperationDescription };
