// igniteui-react-core/esm5/lib/WebFloatPaneActionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebPaneDragActionDescription } from "./WebPaneDragActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebFloatPaneActionDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebFloatPaneActionDescription, _super);
    function WebFloatPaneActionDescription() {
        var _this = _super.call(this) || this;
        _this.s = null;
        _this.k = null;
        _this.o = 0;
        _this.n = 0;
        return _this;
    }
    WebFloatPaneActionDescription.prototype.get_type = function () {
        return "WebFloatPaneAction";
    };
    Object.defineProperty(WebFloatPaneActionDescription.prototype, "actionType", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("ActionType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFloatPaneActionDescription.prototype, "location", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Location");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFloatPaneActionDescription.prototype, "width", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFloatPaneActionDescription.prototype, "height", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    WebFloatPaneActionDescription.$t = markType(WebFloatPaneActionDescription, 'WebFloatPaneActionDescription', WebPaneDragActionDescription.$);
    WebFloatPaneActionDescription.__marshalByValue1 = true;
    WebFloatPaneActionDescription.__marshalByValueAlias1 = "FloatPaneAction";
    return WebFloatPaneActionDescription;
}(WebPaneDragActionDescription));
export { WebFloatPaneActionDescription };
