// igniteui-react-core/esm5/lib/WebFilteringExpressionsTreeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebFilteringExpressionsTreeDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebFilteringExpressionsTreeDescription, _super);
    function WebFilteringExpressionsTreeDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.p = null;
        _this.k = null;
        _this.o = null;
        _this.n = null;
        return _this;
    }
    WebFilteringExpressionsTreeDescription.prototype.get_type = function () {
        return "WebFilteringExpressionsTree";
    };
    Object.defineProperty(WebFilteringExpressionsTreeDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringExpressionsTreeDescription.prototype, "filteringOperands", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("FilteringOperands");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringExpressionsTreeDescription.prototype, "treeType", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("TreeType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringExpressionsTreeDescription.prototype, "owner", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Owner");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringExpressionsTreeDescription.prototype, "operator", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Operator");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringExpressionsTreeDescription.prototype, "fieldName", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("FieldName");
        },
        enumerable: false,
        configurable: true
    });
    WebFilteringExpressionsTreeDescription.$t = markType(WebFilteringExpressionsTreeDescription, 'WebFilteringExpressionsTreeDescription', Description.$);
    WebFilteringExpressionsTreeDescription.__marshalByValue = true;
    WebFilteringExpressionsTreeDescription.__marshalByValueAlias = "FilteringExpressionsTree";
    return WebFilteringExpressionsTreeDescription;
}(Description));
export { WebFilteringExpressionsTreeDescription };
