// igniteui-react-core/esm5/lib/WebFilteringExpressionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebFilteringExpressionDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebFilteringExpressionDescription, _super);
    function WebFilteringExpressionDescription() {
        var _this = _super.call(this) || this;
        _this.p = null;
        _this.h = null;
        _this.m = null;
        _this.k = false;
        return _this;
    }
    WebFilteringExpressionDescription.prototype.get_type = function () {
        return "WebFilteringExpression";
    };
    Object.defineProperty(WebFilteringExpressionDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringExpressionDescription.prototype, "fieldName", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("FieldName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringExpressionDescription.prototype, "condition", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Condition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringExpressionDescription.prototype, "searchVal", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("SearchVal");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFilteringExpressionDescription.prototype, "ignoreCase", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("IgnoreCase");
        },
        enumerable: false,
        configurable: true
    });
    WebFilteringExpressionDescription.$t = markType(WebFilteringExpressionDescription, 'WebFilteringExpressionDescription', Description.$);
    WebFilteringExpressionDescription.__marshalByValue = true;
    WebFilteringExpressionDescription.__marshalByValueAlias = "FilteringExpression";
    return WebFilteringExpressionDescription;
}(Description));
export { WebFilteringExpressionDescription };
