// igniteui-react-core/esm5/lib/WebFloatingPaneResizeMoveEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebFloatingPaneResizeEventArgsDetailDescription } from "./WebFloatingPaneResizeEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebFloatingPaneResizeMoveEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebFloatingPaneResizeMoveEventArgsDetailDescription, _super);
    function WebFloatingPaneResizeMoveEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.w = 0;
        _this.u = 0;
        _this.v = 0;
        _this.t = 0;
        _this.p = null;
        _this.o = null;
        return _this;
    }
    WebFloatingPaneResizeMoveEventArgsDetailDescription.prototype.get_type = function () {
        return "WebFloatingPaneResizeMoveEventArgsDetail";
    };
    Object.defineProperty(WebFloatingPaneResizeMoveEventArgsDetailDescription.prototype, "oldWidth", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("OldWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFloatingPaneResizeMoveEventArgsDetailDescription.prototype, "newWidth", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("NewWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFloatingPaneResizeMoveEventArgsDetailDescription.prototype, "oldHeight", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("OldHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFloatingPaneResizeMoveEventArgsDetailDescription.prototype, "newHeight", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("NewHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFloatingPaneResizeMoveEventArgsDetailDescription.prototype, "oldLocation", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("OldLocation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebFloatingPaneResizeMoveEventArgsDetailDescription.prototype, "newLocation", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("NewLocation");
        },
        enumerable: false,
        configurable: true
    });
    WebFloatingPaneResizeMoveEventArgsDetailDescription.$t = markType(WebFloatingPaneResizeMoveEventArgsDetailDescription, 'WebFloatingPaneResizeMoveEventArgsDetailDescription', WebFloatingPaneResizeEventArgsDetailDescription.$);
    WebFloatingPaneResizeMoveEventArgsDetailDescription.__marshalByValue1 = true;
    WebFloatingPaneResizeMoveEventArgsDetailDescription.__marshalByValueAlias1 = "FloatingPaneResizeMoveEventArgsDetail";
    return WebFloatingPaneResizeMoveEventArgsDetailDescription;
}(WebFloatingPaneResizeEventArgsDetailDescription));
export { WebFloatingPaneResizeMoveEventArgsDetailDescription };
