// igniteui-react-core/esm5/lib/WebGridEditingActionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebGridActionsBaseDirectiveDescription } from "./WebGridActionsBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebGridEditingActionsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebGridEditingActionsDescription, _super);
    function WebGridEditingActionsDescription() {
        var _this = _super.call(this) || this;
        _this.n = false;
        _this.p = false;
        _this.o = false;
        _this.m = false;
        return _this;
    }
    WebGridEditingActionsDescription.prototype.get_type = function () {
        return "WebGridEditingActions";
    };
    Object.defineProperty(WebGridEditingActionsDescription.prototype, "addRow", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("AddRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditingActionsDescription.prototype, "editRow", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("EditRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditingActionsDescription.prototype, "deleteRow", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("DeleteRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditingActionsDescription.prototype, "addChild", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("AddChild");
        },
        enumerable: false,
        configurable: true
    });
    WebGridEditingActionsDescription.$t = markType(WebGridEditingActionsDescription, 'WebGridEditingActionsDescription', WebGridActionsBaseDirectiveDescription.$);
    return WebGridEditingActionsDescription;
}(WebGridActionsBaseDirectiveDescription));
export { WebGridEditingActionsDescription };
