// igniteui-react-core/esm5/lib/WebGridEditDoneEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebGridEditDoneEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebGridEditDoneEventArgsDetailDescription, _super);
    function WebGridEditDoneEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.ab = null;
        _this.z = null;
        _this.ac = null;
        _this.q = null;
        _this.aa = null;
        _this.y = null;
        _this.x = null;
        _this.m = null;
        _this.o = null;
        _this.t = false;
        _this.u = false;
        return _this;
    }
    WebGridEditDoneEventArgsDetailDescription.prototype.get_type = function () {
        return "WebGridEditDoneEventArgsDetail";
    };
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "rowID", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("RowID");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "primaryKey", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("PrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "rowKey", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("RowKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "cellID", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("CellID");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "rowData", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("RowData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "oldValue", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("OldValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "newValue", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("NewValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "column", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Column");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "owner", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Owner");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "isAddRow", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("IsAddRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditDoneEventArgsDetailDescription.prototype, "valid", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Valid");
        },
        enumerable: false,
        configurable: true
    });
    WebGridEditDoneEventArgsDetailDescription.$t = markType(WebGridEditDoneEventArgsDetailDescription, 'WebGridEditDoneEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebGridEditDoneEventArgsDetailDescription.__marshalByValue1 = true;
    WebGridEditDoneEventArgsDetailDescription.__marshalByValueAlias1 = "GridEditDoneEventArgsDetail";
    return WebGridEditDoneEventArgsDetailDescription;
}(WebBaseEventArgsDetailDescription));
export { WebGridEditDoneEventArgsDetailDescription };
