// igniteui-react-core/esm5/lib/WebGridBaseDirectiveDescription_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebGridBaseDirectiveDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebGridBaseDirectiveDescription, _super);
    function WebGridBaseDirectiveDescription() {
        var _this = _super.call(this) || this;
        _this.bi = 0;
        _this.av = false;
        _this.j = null;
        _this.az = false;
        _this.cj = null;
        _this.bq = null;
        _this.cy = null;
        _this.bj = 0;
        _this.ab = null;
        _this.z = null;
        _this.c7 = null;
        _this.dr = null;
        _this.c1 = null;
        _this.k = null;
        _this.h = null;
        _this.cf = null;
        _this.dj = null;
        _this.c6 = null;
        _this.df = null;
        _this.dk = null;
        _this.c9 = null;
        _this.cv = null;
        _this.cu = null;
        _this.ck = null;
        _this.dv = null;
        _this.dw = null;
        _this.dx = null;
        _this.p = null;
        _this.l = null;
        _this.aj = null;
        _this.co = null;
        _this.ae = null;
        _this.ad = null;
        _this.cz = null;
        _this.c0 = null;
        _this.ax = false;
        _this.a0 = false;
        _this.d5 = null;
        _this.a1 = false;
        _this.cx = null;
        _this.d6 = null;
        _this.bh = 0;
        _this.ca = null;
        _this.ci = null;
        _this.ay = false;
        _this.a3 = false;
        _this.ch = null;
        _this.ap = null;
        _this.au = false;
        _this.at = false;
        _this.cp = null;
        _this.d2 = null;
        _this.d1 = null;
        _this.a4 = false;
        _this.ah = null;
        _this.al = null;
        _this.ar = null;
        _this.m = null;
        _this.cw = null;
        _this.dq = null;
        _this.cg = null;
        _this.n = null;
        _this.aw = false;
        _this.bx = null;
        _this.dn = null;
        _this.b6 = null;
        _this.an = null;
        _this.bk = 0;
        _this.a2 = false;
        _this.i = null;
        _this.o = null;
        _this.cn = null;
        _this.br = null;
        _this.ct = null;
        _this.bs = null;
        _this.c8 = null;
        _this.cq = null;
        _this.d4 = null;
        _this.dt = null;
        _this.dp = null;
        _this.b7 = null;
        _this.b2 = null;
        _this.b3 = null;
        _this.bv = null;
        _this.bw = null;
        _this.bt = null;
        _this.bu = null;
        _this.dh = null;
        _this.de = null;
        _this.dg = null;
        _this.di = null;
        _this.by = null;
        _this.b5 = null;
        _this.dy = null;
        _this.dz = null;
        _this.cl = null;
        _this.cm = null;
        _this.c5 = null;
        _this.db = null;
        _this.da = null;
        _this.c4 = null;
        _this.b4 = null;
        _this.cb = null;
        _this.ce = null;
        _this.b9 = null;
        _this.b8 = null;
        _this.b1 = null;
        _this.bz = null;
        _this.b0 = null;
        _this.cs = null;
        _this.dd = null;
        _this.dc = null;
        _this.cr = null;
        _this.du = null;
        _this.ds = null;
        _this.dm = null;
        _this.dl = null;
        _this.bp = null;
        _this.d0 = null;
        _this.d3 = null;
        _this.c2 = null;
        _this.c3 = null;
        _this.cd = null;
        _this.cc = null;
        return _this;
    }
    WebGridBaseDirectiveDescription.prototype.get_type = function () {
        return "WebGridBaseDirective";
    };
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "snackbarDisplayTime", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("SnackbarDisplayTime");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "autoGenerate", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("AutoGenerate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "autoGenerateExclude", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("AutoGenerateExclude");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "moving", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("Moving");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "emptyGridTemplateRef", {
        get: function () {
            return this.cj;
        },
        set: function (a) {
            this.cj = a;
            this.g("EmptyGridTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "addRowEmptyTemplateRef", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            this.bq = a;
            this.g("AddRowEmptyTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "loadingGridTemplateRef", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("LoadingGridTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "summaryRowHeight", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("SummaryRowHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "dataCloneStrategy", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("DataCloneStrategy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "clipboardOptions", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("ClipboardOptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowClassesRef", {
        get: function () {
            return this.c7;
        },
        set: function (a) {
            this.c7 = a;
            this.g("RowClassesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowStylesRef", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("RowStylesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "primaryKey", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.g("PrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnList", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("ColumnList");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "actionStripComponents", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ActionStripComponents");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "dragGhostCustomTemplateRef", {
        get: function () {
            return this.cf;
        },
        set: function (a) {
            this.cf = a;
            this.g("DragGhostCustomTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowEditTextTemplateRef", {
        get: function () {
            return this.dj;
        },
        set: function (a) {
            this.dj = a;
            this.g("RowEditTextTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowAddTextTemplateRef", {
        get: function () {
            return this.c6;
        },
        set: function (a) {
            this.c6 = a;
            this.g("RowAddTextTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowEditActionsTemplateRef", {
        get: function () {
            return this.df;
        },
        set: function (a) {
            this.df = a;
            this.g("RowEditActionsTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowExpandedIndicatorTemplateRef", {
        get: function () {
            return this.dk;
        },
        set: function (a) {
            this.dk = a;
            this.g("RowExpandedIndicatorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowCollapsedIndicatorTemplateRef", {
        get: function () {
            return this.c9;
        },
        set: function (a) {
            this.c9 = a;
            this.g("RowCollapsedIndicatorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "headerExpandedIndicatorTemplateRef", {
        get: function () {
            return this.cv;
        },
        set: function (a) {
            this.cv = a;
            this.g("HeaderExpandedIndicatorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "headerCollapsedIndicatorTemplateRef", {
        get: function () {
            return this.cu;
        },
        set: function (a) {
            this.cu = a;
            this.g("HeaderCollapsedIndicatorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "excelStyleHeaderIconTemplateRef", {
        get: function () {
            return this.ck;
        },
        set: function (a) {
            this.ck = a;
            this.g("ExcelStyleHeaderIconTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "sortAscendingHeaderIconTemplateRef", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.g("SortAscendingHeaderIconTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "sortDescendingHeaderIconTemplateRef", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.g("SortDescendingHeaderIconTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "sortHeaderIconTemplateRef", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("SortHeaderIconTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "toolbar", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Toolbar");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "paginationComponents", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("PaginationComponents");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "resourceStrings", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("ResourceStrings");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "filteringLogic", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.g("FilteringLogic");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "filteringExpressionsTree", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("FilteringExpressionsTree");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "advancedFilteringExpressionsTree", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("AdvancedFilteringExpressionsTree");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "locale", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("Locale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "pagingMode", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("PagingMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "hideRowSelectors", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("HideRowSelectors");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowDraggable", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("RowDraggable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "validationTrigger", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("ValidationTrigger");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowEditable", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("RowEditable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "height", {
        get: function () {
            return this.cx;
        },
        set: function (a) {
            this.cx = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "width", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            this.d6 = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowHeight", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.g("RowHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnWidth", {
        get: function () {
            return this.ca;
        },
        set: function (a) {
            this.ca = a;
            this.g("ColumnWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "emptyGridMessage", {
        get: function () {
            return this.ci;
        },
        set: function (a) {
            this.ci = a;
            this.g("EmptyGridMessage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "isLoading", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("IsLoading");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "shouldGenerate", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("ShouldGenerate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "emptyFilteredGridMessage", {
        get: function () {
            return this.ch;
        },
        set: function (a) {
            this.ch = a;
            this.g("EmptyFilteredGridMessage");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "pinning", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("Pinning");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "allowFiltering", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("AllowFiltering");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "allowAdvancedFiltering", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("AllowAdvancedFiltering");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "filterMode", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("FilterMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "summaryPosition", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            this.d2 = a;
            this.g("SummaryPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "summaryCalculationMode", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            this.d1 = a;
            this.g("SummaryCalculationMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "showSummaryOnCollapse", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("ShowSummaryOnCollapse");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "filterStrategy", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("FilterStrategy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "sortStrategy", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("SortStrategy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "sortingOptions", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("SortingOptions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "selectedRows", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("SelectedRows");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "headSelectorTemplateRef", {
        get: function () {
            return this.cw;
        },
        set: function (a) {
            this.cw = a;
            this.g("HeadSelectorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowSelectorTemplateRef", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("RowSelectorTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "dragIndicatorIconTemplateRef", {
        get: function () {
            return this.cg;
        },
        set: function (a) {
            this.cg = a;
            this.g("DragIndicatorIconTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "sortingExpressions", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("SortingExpressions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "batchEditing", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("BatchEditing");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "cellSelection", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("CellSelection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowSelection", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.g("RowSelection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnSelection", {
        get: function () {
            return this.b6;
        },
        set: function (a) {
            this.b6 = a;
            this.g("ColumnSelection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "outlet", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("Outlet");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "totalRecords", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("TotalRecords");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "selectRowOnClick", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("SelectRowOnClick");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "actualColumnList", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ActualColumnList");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "stateComponents", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("StateComponents");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "filteringExpressionsTreeChangeRef", {
        get: function () {
            return this.cn;
        },
        set: function (a) {
            this.cn = a;
            this.g("FilteringExpressionsTreeChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "advancedFilteringExpressionsTreeChangeRef", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.g("AdvancedFilteringExpressionsTreeChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "gridScrollRef", {
        get: function () {
            return this.ct;
        },
        set: function (a) {
            this.ct = a;
            this.g("GridScrollRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "cellClickRef", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("CellClickRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowClickRef", {
        get: function () {
            return this.c8;
        },
        set: function (a) {
            this.c8 = a;
            this.g("RowClickRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "formGroupCreatedRef", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("FormGroupCreatedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "validationStatusChangeRef", {
        get: function () {
            return this.d4;
        },
        set: function (a) {
            this.d4 = a;
            this.g("ValidationStatusChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "selectedRef", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("SelectedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowSelectionChangingRef", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("RowSelectionChangingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnSelectionChangingRef", {
        get: function () {
            return this.b7;
        },
        set: function (a) {
            this.b7 = a;
            this.g("ColumnSelectionChangingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnPinRef", {
        get: function () {
            return this.b2;
        },
        set: function (a) {
            this.b2 = a;
            this.g("ColumnPinRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnPinnedRef", {
        get: function () {
            return this.b3;
        },
        set: function (a) {
            this.b3 = a;
            this.g("ColumnPinnedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "cellEditEnterRef", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("CellEditEnterRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "cellEditExitRef", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("CellEditExitRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "cellEditRef", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("CellEditRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "cellEditDoneRef", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("CellEditDoneRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowEditEnterRef", {
        get: function () {
            return this.dh;
        },
        set: function (a) {
            this.dh = a;
            this.g("RowEditEnterRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowEditRef", {
        get: function () {
            return this.de;
        },
        set: function (a) {
            this.de = a;
            this.g("RowEditRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowEditDoneRef", {
        get: function () {
            return this.dg;
        },
        set: function (a) {
            this.dg = a;
            this.g("RowEditDoneRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowEditExitRef", {
        get: function () {
            return this.di;
        },
        set: function (a) {
            this.di = a;
            this.g("RowEditExitRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnInitRef", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            this.by = a;
            this.g("ColumnInitRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnsAutogeneratedRef", {
        get: function () {
            return this.b5;
        },
        set: function (a) {
            this.b5 = a;
            this.g("ColumnsAutogeneratedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "sortingRef", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("SortingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "sortingDoneRef", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.g("SortingDoneRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "filteringRef", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            this.cl = a;
            this.g("FilteringRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "filteringDoneRef", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.g("FilteringDoneRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowAddedRef", {
        get: function () {
            return this.c5;
        },
        set: function (a) {
            this.c5 = a;
            this.g("RowAddedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowDeletedRef", {
        get: function () {
            return this.db;
        },
        set: function (a) {
            this.db = a;
            this.g("RowDeletedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowDeleteRef", {
        get: function () {
            return this.da;
        },
        set: function (a) {
            this.da = a;
            this.g("RowDeleteRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowAddRef", {
        get: function () {
            return this.c4;
        },
        set: function (a) {
            this.c4 = a;
            this.g("RowAddRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnResizedRef", {
        get: function () {
            return this.b4;
        },
        set: function (a) {
            this.b4 = a;
            this.g("ColumnResizedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "contextMenuRef", {
        get: function () {
            return this.cb;
        },
        set: function (a) {
            this.cb = a;
            this.g("ContextMenuRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "doubleClickRef", {
        get: function () {
            return this.ce;
        },
        set: function (a) {
            this.ce = a;
            this.g("DoubleClickRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnVisibilityChangingRef", {
        get: function () {
            return this.b9;
        },
        set: function (a) {
            this.b9 = a;
            this.g("ColumnVisibilityChangingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnVisibilityChangedRef", {
        get: function () {
            return this.b8;
        },
        set: function (a) {
            this.b8 = a;
            this.g("ColumnVisibilityChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnMovingStartRef", {
        get: function () {
            return this.b1;
        },
        set: function (a) {
            this.b1 = a;
            this.g("ColumnMovingStartRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnMovingRef", {
        get: function () {
            return this.bz;
        },
        set: function (a) {
            this.bz = a;
            this.g("ColumnMovingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "columnMovingEndRef", {
        get: function () {
            return this.b0;
        },
        set: function (a) {
            this.b0 = a;
            this.g("ColumnMovingEndRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "gridKeydownRef", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("GridKeydownRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowDragStartRef", {
        get: function () {
            return this.dd;
        },
        set: function (a) {
            this.dd = a;
            this.g("RowDragStartRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowDragEndRef", {
        get: function () {
            return this.dc;
        },
        set: function (a) {
            this.dc = a;
            this.g("RowDragEndRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "gridCopyRef", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("GridCopyRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "selectedRowsChangeRef", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.g("SelectedRowsChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowToggleRef", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("RowToggleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowPinningRef", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            this.dm = a;
            this.g("RowPinningRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rowPinnedRef", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            this.dl = a;
            this.g("RowPinnedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "activeNodeChangeRef", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.g("ActiveNodeChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "sortingExpressionsChangeRef", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.g("SortingExpressionsChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "toolbarExportingRef", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.g("ToolbarExportingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "rangeSelectedRef", {
        get: function () {
            return this.c2;
        },
        set: function (a) {
            this.c2 = a;
            this.g("RangeSelectedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "renderedRef", {
        get: function () {
            return this.c3;
        },
        set: function (a) {
            this.c3 = a;
            this.g("RenderedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "dataChangingRef", {
        get: function () {
            return this.cd;
        },
        set: function (a) {
            this.cd = a;
            this.g("DataChangingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridBaseDirectiveDescription.prototype, "dataChangedRef", {
        get: function () {
            return this.cc;
        },
        set: function (a) {
            this.cc = a;
            this.g("DataChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    WebGridBaseDirectiveDescription.$t = markType(WebGridBaseDirectiveDescription, 'WebGridBaseDirectiveDescription', Description.$);
    return WebGridBaseDirectiveDescription;
}(Description));
export { WebGridBaseDirectiveDescription };
/**
 * @hidden
 */
var WebGridToolbarDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebGridToolbarDescription, _super);
    function WebGridToolbarDescription() {
        var _this = _super.call(this) || this;
        _this.l = false;
        _this.j = null;
        _this.n = null;
        _this.h = null;
        return _this;
    }
    WebGridToolbarDescription.prototype.get_type = function () {
        return "WebGridToolbar";
    };
    Object.defineProperty(WebGridToolbarDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridToolbarDescription.prototype, "showProgress", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ShowProgress");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridToolbarDescription.prototype, "grid", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Grid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridToolbarDescription.prototype, "name", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridToolbarDescription.prototype, "tools", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Tools");
        },
        enumerable: false,
        configurable: true
    });
    WebGridToolbarDescription.$t = markType(WebGridToolbarDescription, 'WebGridToolbarDescription', Description.$);
    return WebGridToolbarDescription;
}(Description));
export { WebGridToolbarDescription };
