// igniteui-react-core/esm5/lib/WebGridEditEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebGridEditEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebGridEditEventArgsDetailDescription, _super);
    function WebGridEditEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.o = false;
        _this.y = null;
        _this.w = null;
        _this.z = null;
        _this.l = null;
        _this.x = null;
        _this.v = null;
        _this.u = null;
        _this.h = null;
        _this.j = null;
        _this.p = false;
        _this.q = false;
        return _this;
    }
    WebGridEditEventArgsDetailDescription.prototype.get_type = function () {
        return "WebGridEditEventArgsDetail";
    };
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "cancel", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Cancel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "rowID", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("RowID");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "primaryKey", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("PrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "rowKey", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("RowKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "cellID", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("CellID");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "rowData", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("RowData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "oldValue", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("OldValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "newValue", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("NewValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "column", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Column");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "owner", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Owner");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "isAddRow", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("IsAddRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridEditEventArgsDetailDescription.prototype, "valid", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Valid");
        },
        enumerable: false,
        configurable: true
    });
    WebGridEditEventArgsDetailDescription.$t = markType(WebGridEditEventArgsDetailDescription, 'WebGridEditEventArgsDetailDescription', Description.$);
    WebGridEditEventArgsDetailDescription.__marshalByValue = true;
    WebGridEditEventArgsDetailDescription.__marshalByValueAlias = "GridEditEventArgsDetail";
    return WebGridEditEventArgsDetailDescription;
}(Description));
export { WebGridEditEventArgsDetailDescription };
