// igniteui-react-core/esm5/lib/WebGroupByRecordDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebGroupByRecordDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebGroupByRecordDescription, _super);
    function WebGroupByRecordDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.q = 0;
        _this.v = null;
        _this.t = null;
        _this.l = null;
        _this.h = null;
        _this.p = 0;
        _this.j = null;
        return _this;
    }
    WebGroupByRecordDescription.prototype.get_type = function () {
        return "WebGroupByRecord";
    };
    Object.defineProperty(WebGroupByRecordDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGroupByRecordDescription.prototype, "expression", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Expression");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGroupByRecordDescription.prototype, "level", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Level");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGroupByRecordDescription.prototype, "recordsRef", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("RecordsRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGroupByRecordDescription.prototype, "value", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGroupByRecordDescription.prototype, "groupParent", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("GroupParent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGroupByRecordDescription.prototype, "groups", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Groups");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGroupByRecordDescription.prototype, "height", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGroupByRecordDescription.prototype, "column", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Column");
        },
        enumerable: false,
        configurable: true
    });
    WebGroupByRecordDescription.$t = markType(WebGroupByRecordDescription, 'WebGroupByRecordDescription', Description.$);
    return WebGroupByRecordDescription;
}(Description));
export { WebGroupByRecordDescription };
