// igniteui-react-core/esm5/lib/WebGridKeydownEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebGridKeydownEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebGridKeydownEventArgsDetailDescription, _super);
    function WebGridKeydownEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.u = null;
        _this.q = null;
        _this.p = null;
        _this.n = false;
        return _this;
    }
    WebGridKeydownEventArgsDetailDescription.prototype.get_type = function () {
        return "WebGridKeydownEventArgsDetail";
    };
    Object.defineProperty(WebGridKeydownEventArgsDetailDescription.prototype, "targetType", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("TargetType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridKeydownEventArgsDetailDescription.prototype, "target", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Target");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridKeydownEventArgsDetailDescription.prototype, "event", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Event");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridKeydownEventArgsDetailDescription.prototype, "cancel", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Cancel");
        },
        enumerable: false,
        configurable: true
    });
    WebGridKeydownEventArgsDetailDescription.$t = markType(WebGridKeydownEventArgsDetailDescription, 'WebGridKeydownEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebGridKeydownEventArgsDetailDescription.__marshalByValue1 = true;
    WebGridKeydownEventArgsDetailDescription.__marshalByValueAlias1 = "GridKeydownEventArgsDetail";
    return WebGridKeydownEventArgsDetailDescription;
}(WebBaseEventArgsDetailDescription));
export { WebGridKeydownEventArgsDetailDescription };
