// igniteui-react-core/esm5/lib/WebGridStateInfoDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebGridStateInfoDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebGridStateInfoDetailDescription, _super);
    function WebGridStateInfoDetailDescription() {
        var _this = _super.call(this) || this;
        _this.i = null;
        _this.y = null;
        _this.x = null;
        _this.ad = null;
        _this.ak = false;
        _this.o = null;
        _this.ab = null;
        _this.h = null;
        _this.n = null;
        _this.j = null;
        _this.m = null;
        _this.af = null;
        _this.k = null;
        _this.l = null;
        _this.an = null;
        _this.ah = null;
        return _this;
    }
    WebGridStateInfoDetailDescription.prototype.get_type = function () {
        return "WebGridStateInfoDetail";
    };
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "columns", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Columns");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "filtering", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("Filtering");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "advancedFiltering", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("AdvancedFiltering");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "paging", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("Paging");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "moving", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("Moving");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "sorting", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Sorting");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "groupBy", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("GroupBy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "cellSelection", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("CellSelection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "rowSelection", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("RowSelection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "columnSelection", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ColumnSelection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "rowPinning", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("RowPinning");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "pinningConfig", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("PinningConfig");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "expansion", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Expansion");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "rowIslands", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("RowIslands");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "id", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridStateInfoDetailDescription.prototype, "pivotConfiguration", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("PivotConfiguration");
        },
        enumerable: false,
        configurable: true
    });
    WebGridStateInfoDetailDescription.$t = markType(WebGridStateInfoDetailDescription, 'WebGridStateInfoDetailDescription', Description.$);
    WebGridStateInfoDetailDescription.__marshalByValue = true;
    WebGridStateInfoDetailDescription.__marshalByValueAlias = "GridStateInfoDetail";
    return WebGridStateInfoDetailDescription;
}(Description));
export { WebGridStateInfoDetailDescription };
