// igniteui-react-core/esm5/lib/WebGridResourceStringsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebGridResourceStringsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebGridResourceStringsDescription, _super);
    function WebGridResourceStringsDescription() {
        var _this = _super.call(this) || this;
        _this.co = null;
        _this.cp = null;
        _this.cn = null;
        _this.ah = null;
        _this.ai = null;
        _this.bi = null;
        _this.cc = null;
        _this.cf = null;
        _this.ce = null;
        _this.cb = null;
        _this.cd = null;
        _this.cg = null;
        _this.b9 = null;
        _this.af = null;
        _this.ca = null;
        _this.bq = null;
        _this.br = null;
        _this.ch = null;
        _this.bu = null;
        _this.bv = null;
        _this.bs = null;
        _this.bt = null;
        _this.b6 = null;
        _this.b8 = null;
        _this.b7 = null;
        _this.bo = null;
        _this.bj = null;
        _this.bl = null;
        _this.b5 = null;
        _this.bn = null;
        _this.bm = null;
        _this.ck = null;
        _this.cm = null;
        _this.ci = null;
        _this.bz = null;
        _this.b3 = null;
        _this.cj = null;
        _this.b0 = null;
        _this.b4 = null;
        _this.bx = null;
        _this.b1 = null;
        _this.by = null;
        _this.b2 = null;
        _this.cl = null;
        _this.bw = null;
        _this.bk = null;
        _this.bp = null;
        _this.dw = null;
        _this.d0 = null;
        _this.dz = null;
        _this.d1 = null;
        _this.dv = null;
        _this.dx = null;
        _this.dy = null;
        _this.ax = null;
        _this.ay = null;
        _this.az = null;
        _this.a0 = null;
        _this.aw = null;
        _this.a1 = null;
        _this.a2 = null;
        _this.a3 = null;
        _this.a4 = null;
        _this.a5 = null;
        _this.av = null;
        _this.ap = null;
        _this.aq = null;
        _this.ar = null;
        _this.an = null;
        _this.ak = null;
        _this.bb = null;
        _this.bd = null;
        _this.be = null;
        _this.aj = null;
        _this.al = null;
        _this.a6 = null;
        _this.bf = null;
        _this.ba = null;
        _this.bh = null;
        _this.bc = null;
        _this.au = null;
        _this.bg = null;
        _this.a9 = null;
        _this.at = null;
        _this.am = null;
        _this.ao = null;
        _this.as = null;
        _this.a8 = null;
        _this.a7 = null;
        _this.ac = null;
        _this.r = null;
        _this.aa = null;
        _this.y = null;
        _this.u = null;
        _this.v = null;
        _this.s = null;
        _this.ab = null;
        _this.q = null;
        _this.ad = null;
        _this.w = null;
        _this.x = null;
        _this.z = null;
        _this.t = null;
        _this.ae = null;
        _this.cx = null;
        _this.cq = null;
        _this.cr = null;
        _this.cy = null;
        _this.cz = null;
        _this.dr = null;
        _this.dq = null;
        _this.ds = null;
        _this.d2 = null;
        _this.eb = null;
        _this.d9 = null;
        _this.ec = null;
        _this.ea = null;
        _this.d4 = null;
        _this.d3 = null;
        _this.d6 = null;
        _this.d5 = null;
        _this.d8 = null;
        _this.d7 = null;
        _this.du = null;
        _this.dt = null;
        _this.k = null;
        _this.i = null;
        _this.p = null;
        _this.h = null;
        _this.j = null;
        _this.n = null;
        _this.o = null;
        _this.m = null;
        _this.l = null;
        _this.da = null;
        _this.c1 = null;
        _this.c5 = null;
        _this.c4 = null;
        _this.c6 = null;
        _this.c0 = null;
        _this.c3 = null;
        _this.c2 = null;
        _this.c8 = null;
        _this.c7 = null;
        _this.de = null;
        _this.db = null;
        _this.dc = null;
        _this.df = null;
        _this.dh = null;
        _this.c9 = null;
        _this.dg = null;
        _this.dn = null;
        _this.dd = null;
        _this.dj = null;
        _this.dl = null;
        _this.di = null;
        _this.dm = null;
        _this.dk = null;
        _this.dp = null;
        _this.cv = null;
        _this.ct = null;
        _this.cu = null;
        _this.cs = null;
        _this.ag = null;
        _this.cw = null;
        return _this;
    }
    WebGridResourceStringsDescription.prototype.get_type = function () {
        return "WebGridResourceStrings";
    };
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_groupByArea_message", {
        get: function () {
            return this.co;
        },
        set: function (a) {
            this.co = a;
            this.g("Igx_grid_groupByArea_message");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_groupByArea_select_message", {
        get: function () {
            return this.cp;
        },
        set: function (a) {
            this.cp = a;
            this.g("Igx_grid_groupByArea_select_message");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_groupByArea_deselect_message", {
        get: function () {
            return this.cn;
        },
        set: function (a) {
            this.cn = a;
            this.g("Igx_grid_groupByArea_deselect_message");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_emptyFilteredGrid_message", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("Igx_grid_emptyFilteredGrid_message");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_emptyGrid_message", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("Igx_grid_emptyGrid_message");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("Igx_grid_filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_row_close", {
        get: function () {
            return this.cc;
        },
        set: function (a) {
            this.cc = a;
            this.g("Igx_grid_filter_row_close");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_row_reset", {
        get: function () {
            return this.cf;
        },
        set: function (a) {
            this.cf = a;
            this.g("Igx_grid_filter_row_reset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_row_placeholder", {
        get: function () {
            return this.ce;
        },
        set: function (a) {
            this.ce = a;
            this.g("Igx_grid_filter_row_placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_row_boolean_placeholder", {
        get: function () {
            return this.cb;
        },
        set: function (a) {
            this.cb = a;
            this.g("Igx_grid_filter_row_boolean_placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_row_date_placeholder", {
        get: function () {
            return this.cd;
        },
        set: function (a) {
            this.cd = a;
            this.g("Igx_grid_filter_row_date_placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_row_time_placeholder", {
        get: function () {
            return this.cg;
        },
        set: function (a) {
            this.cg = a;
            this.g("Igx_grid_filter_row_time_placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_operator_and", {
        get: function () {
            return this.b9;
        },
        set: function (a) {
            this.b9 = a;
            this.g("Igx_grid_filter_operator_and");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_complex_filter", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("Igx_grid_complex_filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_operator_or", {
        get: function () {
            return this.ca;
        },
        set: function (a) {
            this.ca = a;
            this.g("Igx_grid_filter_operator_or");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_contains", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            this.bq = a;
            this.g("Igx_grid_filter_contains");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_doesNotContain", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.g("Igx_grid_filter_doesNotContain");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_startsWith", {
        get: function () {
            return this.ch;
        },
        set: function (a) {
            this.ch = a;
            this.g("Igx_grid_filter_startsWith");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_endsWith", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("Igx_grid_filter_endsWith");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_equals", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("Igx_grid_filter_equals");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_doesNotEqual", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("Igx_grid_filter_doesNotEqual");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_empty", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("Igx_grid_filter_empty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_notEmpty", {
        get: function () {
            return this.b6;
        },
        set: function (a) {
            this.b6 = a;
            this.g("Igx_grid_filter_notEmpty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_null", {
        get: function () {
            return this.b8;
        },
        set: function (a) {
            this.b8 = a;
            this.g("Igx_grid_filter_null");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_notNull", {
        get: function () {
            return this.b7;
        },
        set: function (a) {
            this.b7 = a;
            this.g("Igx_grid_filter_notNull");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_before", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            this.bo = a;
            this.g("Igx_grid_filter_before");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_after", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("Igx_grid_filter_after");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_at", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.g("Igx_grid_filter_at");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_not_at", {
        get: function () {
            return this.b5;
        },
        set: function (a) {
            this.b5 = a;
            this.g("Igx_grid_filter_not_at");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_at_before", {
        get: function () {
            return this.bn;
        },
        set: function (a) {
            this.bn = a;
            this.g("Igx_grid_filter_at_before");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_at_after", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            this.bm = a;
            this.g("Igx_grid_filter_at_after");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_today", {
        get: function () {
            return this.ck;
        },
        set: function (a) {
            this.ck = a;
            this.g("Igx_grid_filter_today");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_yesterday", {
        get: function () {
            return this.cm;
        },
        set: function (a) {
            this.cm = a;
            this.g("Igx_grid_filter_yesterday");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_thisMonth", {
        get: function () {
            return this.ci;
        },
        set: function (a) {
            this.ci = a;
            this.g("Igx_grid_filter_thisMonth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_lastMonth", {
        get: function () {
            return this.bz;
        },
        set: function (a) {
            this.bz = a;
            this.g("Igx_grid_filter_lastMonth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_nextMonth", {
        get: function () {
            return this.b3;
        },
        set: function (a) {
            this.b3 = a;
            this.g("Igx_grid_filter_nextMonth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_thisYear", {
        get: function () {
            return this.cj;
        },
        set: function (a) {
            this.cj = a;
            this.g("Igx_grid_filter_thisYear");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_lastYear", {
        get: function () {
            return this.b0;
        },
        set: function (a) {
            this.b0 = a;
            this.g("Igx_grid_filter_lastYear");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_nextYear", {
        get: function () {
            return this.b4;
        },
        set: function (a) {
            this.b4 = a;
            this.g("Igx_grid_filter_nextYear");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_greaterThan", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("Igx_grid_filter_greaterThan");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_lessThan", {
        get: function () {
            return this.b1;
        },
        set: function (a) {
            this.b1 = a;
            this.g("Igx_grid_filter_lessThan");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_greaterThanOrEqualTo", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            this.by = a;
            this.g("Igx_grid_filter_greaterThanOrEqualTo");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_lessThanOrEqualTo", {
        get: function () {
            return this.b2;
        },
        set: function (a) {
            this.b2 = a;
            this.g("Igx_grid_filter_lessThanOrEqualTo");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_true", {
        get: function () {
            return this.cl;
        },
        set: function (a) {
            this.cl = a;
            this.g("Igx_grid_filter_true");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_false", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("Igx_grid_filter_false");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_all", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("Igx_grid_filter_all");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_filter_condition_placeholder", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.g("Igx_grid_filter_condition_placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_summary_count", {
        get: function () {
            return this.dw;
        },
        set: function (a) {
            this.dw = a;
            this.g("Igx_grid_summary_count");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_summary_min", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.g("Igx_grid_summary_min");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_summary_max", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.g("Igx_grid_summary_max");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_summary_sum", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            this.d1 = a;
            this.g("Igx_grid_summary_sum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_summary_average", {
        get: function () {
            return this.dv;
        },
        set: function (a) {
            this.dv = a;
            this.g("Igx_grid_summary_average");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_summary_earliest", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.g("Igx_grid_summary_earliest");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_summary_latest", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.g("Igx_grid_summary_latest");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_moving_left", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("Igx_grid_excel_filter_moving_left");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_moving_left_short", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("Igx_grid_excel_filter_moving_left_short");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_moving_right", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("Igx_grid_excel_filter_moving_right");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_moving_right_short", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("Igx_grid_excel_filter_moving_right_short");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_moving_header", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("Igx_grid_excel_filter_moving_header");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_sorting_asc", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.g("Igx_grid_excel_filter_sorting_asc");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_sorting_asc_short", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.g("Igx_grid_excel_filter_sorting_asc_short");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_sorting_desc", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("Igx_grid_excel_filter_sorting_desc");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_sorting_desc_short", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.g("Igx_grid_excel_filter_sorting_desc_short");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_sorting_header", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("Igx_grid_excel_filter_sorting_header");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_filter_clear", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("Igx_grid_excel_filter_clear");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_custom_dialog_add", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("Igx_grid_excel_custom_dialog_add");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_custom_dialog_clear", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("Igx_grid_excel_custom_dialog_clear");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_custom_dialog_header", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("Igx_grid_excel_custom_dialog_header");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_cancel", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("Igx_grid_excel_cancel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_apply", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("Igx_grid_excel_apply");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_search_placeholder", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("Igx_grid_excel_search_placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_select_all", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("Igx_grid_excel_select_all");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_select_all_search_results", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.g("Igx_grid_excel_select_all_search_results");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_add_to_filter", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("Igx_grid_excel_add_to_filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_blanks", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("Igx_grid_excel_blanks");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_hide", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("Igx_grid_excel_hide");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_show", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("Igx_grid_excel_show");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_pin", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("Igx_grid_excel_pin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_unpin", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.g("Igx_grid_excel_unpin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_select", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("Igx_grid_excel_select");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_deselect", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("Igx_grid_excel_deselect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_text_filter", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.g("Igx_grid_excel_text_filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_number_filter", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("Igx_grid_excel_number_filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_date_filter", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("Igx_grid_excel_date_filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_boolean_filter", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("Igx_grid_excel_boolean_filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_currency_filter", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("Igx_grid_excel_currency_filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_custom_filter", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("Igx_grid_excel_custom_filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_no_matches", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("Igx_grid_excel_no_matches");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_excel_matches_count", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("Igx_grid_excel_matches_count");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_title", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("Igx_grid_advanced_filter_title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_and_group", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Igx_grid_advanced_filter_and_group");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_or_group", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("Igx_grid_advanced_filter_or_group");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_end_group", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("Igx_grid_advanced_filter_end_group");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_create_and_group", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Igx_grid_advanced_filter_create_and_group");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_create_or_group", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("Igx_grid_advanced_filter_create_or_group");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_and_label", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Igx_grid_advanced_filter_and_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_or_label", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("Igx_grid_advanced_filter_or_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_add_condition", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Igx_grid_advanced_filter_add_condition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_ungroup", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("Igx_grid_advanced_filter_ungroup");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_delete", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("Igx_grid_advanced_filter_delete");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_delete_filters", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("Igx_grid_advanced_filter_delete_filters");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_initial_text", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("Igx_grid_advanced_filter_initial_text");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_column_placeholder", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("Igx_grid_advanced_filter_column_placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_advanced_filter_value_placeholder", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("Igx_grid_advanced_filter_value_placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pinned_row_indicator", {
        get: function () {
            return this.cx;
        },
        set: function (a) {
            this.cx = a;
            this.g("Igx_grid_pinned_row_indicator");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_hiding_check_all_label", {
        get: function () {
            return this.cq;
        },
        set: function (a) {
            this.cq = a;
            this.g("Igx_grid_hiding_check_all_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_hiding_uncheck_all_label", {
        get: function () {
            return this.cr;
        },
        set: function (a) {
            this.cr = a;
            this.g("Igx_grid_hiding_uncheck_all_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pinning_check_all_label", {
        get: function () {
            return this.cy;
        },
        set: function (a) {
            this.cy = a;
            this.g("Igx_grid_pinning_check_all_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pinning_uncheck_all_label", {
        get: function () {
            return this.cz;
        },
        set: function (a) {
            this.cz = a;
            this.g("Igx_grid_pinning_uncheck_all_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_row_edit_btn_done", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("Igx_grid_row_edit_btn_done");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_row_edit_btn_cancel", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("Igx_grid_row_edit_btn_cancel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_row_edit_text", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("Igx_grid_row_edit_text");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_actions_filter_prompt", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            this.d2 = a;
            this.g("Igx_grid_toolbar_actions_filter_prompt");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_pinning_button_tooltip", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.g("Igx_grid_toolbar_pinning_button_tooltip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_hiding_button_tooltip", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.g("Igx_grid_toolbar_hiding_button_tooltip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_pinning_title", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            this.ec = a;
            this.g("Igx_grid_toolbar_pinning_title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_hiding_title", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.g("Igx_grid_toolbar_hiding_title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_advanced_filtering_button_tooltip", {
        get: function () {
            return this.d4;
        },
        set: function (a) {
            this.d4 = a;
            this.g("Igx_grid_toolbar_advanced_filtering_button_tooltip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_advanced_filtering_button_label", {
        get: function () {
            return this.d3;
        },
        set: function (a) {
            this.d3 = a;
            this.g("Igx_grid_toolbar_advanced_filtering_button_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_exporter_button_tooltip", {
        get: function () {
            return this.d6;
        },
        set: function (a) {
            this.d6 = a;
            this.g("Igx_grid_toolbar_exporter_button_tooltip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_exporter_button_label", {
        get: function () {
            return this.d5;
        },
        set: function (a) {
            this.d5 = a;
            this.g("Igx_grid_toolbar_exporter_button_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_exporter_excel_entry_text", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.g("Igx_grid_toolbar_exporter_excel_entry_text");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_toolbar_exporter_csv_entry_text", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.g("Igx_grid_toolbar_exporter_csv_entry_text");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_snackbar_addrow_label", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.g("Igx_grid_snackbar_addrow_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_snackbar_addrow_actiontext", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.g("Igx_grid_snackbar_addrow_actiontext");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_actions_edit_label", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Igx_grid_actions_edit_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_actions_add_label", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Igx_grid_actions_add_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_add_row_label", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Igx_grid_add_row_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_actions_add_child_label", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Igx_grid_actions_add_child_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_actions_delete_label", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Igx_grid_actions_delete_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_actions_pin_label", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Igx_grid_actions_pin_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_actions_unpin_label", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Igx_grid_actions_unpin_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_actions_jumpUp_label", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Igx_grid_actions_jumpUp_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_actions_jumpDown_label", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Igx_grid_actions_jumpDown_label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_date_dimension_total", {
        get: function () {
            return this.da;
        },
        set: function (a) {
            this.da = a;
            this.g("Igx_grid_pivot_date_dimension_total");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_aggregate_count", {
        get: function () {
            return this.c1;
        },
        set: function (a) {
            this.c1 = a;
            this.g("Igx_grid_pivot_aggregate_count");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_aggregate_min", {
        get: function () {
            return this.c5;
        },
        set: function (a) {
            this.c5 = a;
            this.g("Igx_grid_pivot_aggregate_min");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_aggregate_max", {
        get: function () {
            return this.c4;
        },
        set: function (a) {
            this.c4 = a;
            this.g("Igx_grid_pivot_aggregate_max");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_aggregate_sum", {
        get: function () {
            return this.c6;
        },
        set: function (a) {
            this.c6 = a;
            this.g("Igx_grid_pivot_aggregate_sum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_aggregate_avg", {
        get: function () {
            return this.c0;
        },
        set: function (a) {
            this.c0 = a;
            this.g("Igx_grid_pivot_aggregate_avg");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_aggregate_date_latest", {
        get: function () {
            return this.c3;
        },
        set: function (a) {
            this.c3 = a;
            this.g("Igx_grid_pivot_aggregate_date_latest");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_aggregate_date_earliest", {
        get: function () {
            return this.c2;
        },
        set: function (a) {
            this.c2 = a;
            this.g("Igx_grid_pivot_aggregate_date_earliest");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_aggregate_time_latest", {
        get: function () {
            return this.c8;
        },
        set: function (a) {
            this.c8 = a;
            this.g("Igx_grid_pivot_aggregate_time_latest");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_aggregate_time_earliest", {
        get: function () {
            return this.c7;
        },
        set: function (a) {
            this.c7 = a;
            this.g("Igx_grid_pivot_aggregate_time_earliest");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_empty_row_drop_area", {
        get: function () {
            return this.de;
        },
        set: function (a) {
            this.de = a;
            this.g("Igx_grid_pivot_empty_row_drop_area");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_empty_column_drop_area", {
        get: function () {
            return this.db;
        },
        set: function (a) {
            this.db = a;
            this.g("Igx_grid_pivot_empty_column_drop_area");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_empty_filter_drop_area", {
        get: function () {
            return this.dc;
        },
        set: function (a) {
            this.dc = a;
            this.g("Igx_grid_pivot_empty_filter_drop_area");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_empty_value_drop_area", {
        get: function () {
            return this.df;
        },
        set: function (a) {
            this.df = a;
            this.g("Igx_grid_pivot_empty_value_drop_area");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_row_drop_chip", {
        get: function () {
            return this.dh;
        },
        set: function (a) {
            this.dh = a;
            this.g("Igx_grid_pivot_row_drop_chip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_column_drop_chip", {
        get: function () {
            return this.c9;
        },
        set: function (a) {
            this.c9 = a;
            this.g("Igx_grid_pivot_column_drop_chip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_filter_drop_chip", {
        get: function () {
            return this.dg;
        },
        set: function (a) {
            this.dg = a;
            this.g("Igx_grid_pivot_filter_drop_chip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_value_drop_chip", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.g("Igx_grid_pivot_value_drop_chip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_empty_message", {
        get: function () {
            return this.dd;
        },
        set: function (a) {
            this.dd = a;
            this.g("Igx_grid_pivot_empty_message");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_selector_filters", {
        get: function () {
            return this.dj;
        },
        set: function (a) {
            this.dj = a;
            this.g("Igx_grid_pivot_selector_filters");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_selector_rows", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            this.dl = a;
            this.g("Igx_grid_pivot_selector_rows");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_selector_columns", {
        get: function () {
            return this.di;
        },
        set: function (a) {
            this.di = a;
            this.g("Igx_grid_pivot_selector_columns");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_selector_values", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            this.dm = a;
            this.g("Igx_grid_pivot_selector_values");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pivot_selector_panel_empty", {
        get: function () {
            return this.dk;
        },
        set: function (a) {
            this.dk = a;
            this.g("Igx_grid_pivot_selector_panel_empty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_required_validation_error", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("Igx_grid_required_validation_error");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_min_validation_error", {
        get: function () {
            return this.cv;
        },
        set: function (a) {
            this.cv = a;
            this.g("Igx_grid_min_validation_error");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_max_validation_error", {
        get: function () {
            return this.ct;
        },
        set: function (a) {
            this.ct = a;
            this.g("Igx_grid_max_validation_error");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_min_length_validation_error", {
        get: function () {
            return this.cu;
        },
        set: function (a) {
            this.cu = a;
            this.g("Igx_grid_min_length_validation_error");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_max_length_validation_error", {
        get: function () {
            return this.cs;
        },
        set: function (a) {
            this.cs = a;
            this.g("Igx_grid_max_length_validation_error");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_email_validation_error", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("Igx_grid_email_validation_error");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebGridResourceStringsDescription.prototype, "igx_grid_pattern_validation_error", {
        get: function () {
            return this.cw;
        },
        set: function (a) {
            this.cw = a;
            this.g("Igx_grid_pattern_validation_error");
        },
        enumerable: false,
        configurable: true
    });
    WebGridResourceStringsDescription.$t = markType(WebGridResourceStringsDescription, 'WebGridResourceStringsDescription', Description.$);
    return WebGridResourceStringsDescription;
}(Description));
export { WebGridResourceStringsDescription };
