// igniteui-react-core/esm5/lib/WebHierarchicalGridDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebHierarchicalGridBaseDirectiveDescription } from "./WebHierarchicalGridBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebHierarchicalGridDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebHierarchicalGridDescription, _super);
    function WebHierarchicalGridDescription() {
        var _this = _super.call(this) || this;
        _this.gx = null;
        _this.g4 = null;
        _this.g3 = null;
        _this.g1 = 0;
        _this.gz = false;
        return _this;
    }
    WebHierarchicalGridDescription.prototype.get_type = function () {
        return "WebHierarchicalGrid";
    };
    Object.defineProperty(WebHierarchicalGridDescription.prototype, "childLayoutList", {
        get: function () {
            return this.gx;
        },
        set: function (a) {
            this.gx = a;
            this.g("ChildLayoutList");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHierarchicalGridDescription.prototype, "id", {
        get: function () {
            return this.g4;
        },
        set: function (a) {
            this.g4 = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHierarchicalGridDescription.prototype, "dataRef", {
        get: function () {
            return this.g3;
        },
        set: function (a) {
            this.g3 = a;
            this.g("DataRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHierarchicalGridDescription.prototype, "totalItemCount", {
        get: function () {
            return this.g1;
        },
        set: function (a) {
            this.g1 = a;
            this.g("TotalItemCount");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHierarchicalGridDescription.prototype, "expandChildren", {
        get: function () {
            return this.gz;
        },
        set: function (a) {
            this.gz = a;
            this.g("ExpandChildren");
        },
        enumerable: false,
        configurable: true
    });
    WebHierarchicalGridDescription.$t = markType(WebHierarchicalGridDescription, 'WebHierarchicalGridDescription', WebHierarchicalGridBaseDirectiveDescription.$);
    return WebHierarchicalGridDescription;
}(WebHierarchicalGridBaseDirectiveDescription));
export { WebHierarchicalGridDescription };
