// igniteui-react-core/esm5/lib/WebHeaderTypeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebHeaderTypeDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebHeaderTypeDescription, _super);
    function WebHeaderTypeDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.l = false;
        _this.j = false;
        _this.k = false;
        _this.m = false;
        _this.r = null;
        return _this;
    }
    WebHeaderTypeDescription.prototype.get_type = function () {
        return "WebHeaderType";
    };
    Object.defineProperty(WebHeaderTypeDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHeaderTypeDescription.prototype, "column", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Column");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHeaderTypeDescription.prototype, "sorted", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Sorted");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHeaderTypeDescription.prototype, "selectable", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Selectable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHeaderTypeDescription.prototype, "selected", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Selected");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHeaderTypeDescription.prototype, "title", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Title");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHeaderTypeDescription.prototype, "sortDirection", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("SortDirection");
        },
        enumerable: false,
        configurable: true
    });
    WebHeaderTypeDescription.$t = markType(WebHeaderTypeDescription, 'WebHeaderTypeDescription', Description.$);
    return WebHeaderTypeDescription;
}(Description));
export { WebHeaderTypeDescription };
