// igniteui-react-core/esm5/lib/WebInputBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebInputBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebInputBaseDescription, _super);
    function WebInputBaseDescription() {
        var _this = _super.call(this) || this;
        _this.j = false;
        _this.k = false;
        _this.y = null;
        _this.w = null;
        _this.h = false;
        _this.l = false;
        _this.r = null;
        _this.x = null;
        _this.i = false;
        _this.v = null;
        _this.u = null;
        _this.t = null;
        return _this;
    }
    WebInputBaseDescription.prototype.get_type = function () {
        return "WebInputBase";
    };
    Object.defineProperty(WebInputBaseDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "outlined", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Outlined");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "readOnly", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("ReadOnly");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "placeholder", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("Placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "label", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("Label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "disabled", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "required", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Required");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "defaultValue", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("DefaultValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "name", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "invalid", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Invalid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "inputOcurredRef", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("InputOcurredRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "focusRef", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("FocusRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "blurRef", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("BlurRef");
        },
        enumerable: false,
        configurable: true
    });
    WebInputBaseDescription.$t = markType(WebInputBaseDescription, 'WebInputBaseDescription', Description.$);
    return WebInputBaseDescription;
}(Description));
export { WebInputBaseDescription };
