// igniteui-react-core/esm5/lib/WebInputDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebInputBaseDescription } from "./WebInputBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebInputDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebInputDescription, _super);
    function WebInputDescription() {
        var _this = _super.call(this) || this;
        _this.ax = null;
        _this.as = null;
        _this.at = null;
        _this.aw = null;
        _this.al = 0;
        _this.ak = 0;
        _this.av = null;
        _this.au = null;
        _this.am = 0;
        _this.ag = false;
        _this.aq = null;
        _this.ah = false;
        _this.ar = null;
        return _this;
    }
    WebInputDescription.prototype.get_type = function () {
        return "WebInput";
    };
    Object.defineProperty(WebInputDescription.prototype, "value", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "displayType", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("DisplayType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "inputMode", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.g("InputMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "pattern", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("Pattern");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "minLength", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("MinLength");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "maxLength", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("MaxLength");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "min", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("Min");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "max", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("Max");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "step", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("Step");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "autofocus", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("Autofocus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "autocomplete", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("Autocomplete");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "validateOnly", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("ValidateOnly");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebInputDescription.prototype, "changeRef", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    WebInputDescription.$t = markType(WebInputDescription, 'WebInputDescription', WebInputBaseDescription.$);
    return WebInputDescription;
}(WebInputBaseDescription));
export { WebInputDescription };
