// igniteui-react-core/esm5/lib/WebPinColumnCancellableEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebPinColumnCancellableEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebPinColumnCancellableEventArgsDetailDescription, _super);
    function WebPinColumnCancellableEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.o = 0;
        _this.l = false;
        _this.k = false;
        return _this;
    }
    WebPinColumnCancellableEventArgsDetailDescription.prototype.get_type = function () {
        return "WebPinColumnCancellableEventArgsDetail";
    };
    Object.defineProperty(WebPinColumnCancellableEventArgsDetailDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPinColumnCancellableEventArgsDetailDescription.prototype, "column", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Column");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPinColumnCancellableEventArgsDetailDescription.prototype, "insertAtIndex", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("InsertAtIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPinColumnCancellableEventArgsDetailDescription.prototype, "isPinned", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("IsPinned");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPinColumnCancellableEventArgsDetailDescription.prototype, "cancel", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Cancel");
        },
        enumerable: false,
        configurable: true
    });
    WebPinColumnCancellableEventArgsDetailDescription.$t = markType(WebPinColumnCancellableEventArgsDetailDescription, 'WebPinColumnCancellableEventArgsDetailDescription', Description.$);
    WebPinColumnCancellableEventArgsDetailDescription.__marshalByValue = true;
    WebPinColumnCancellableEventArgsDetailDescription.__marshalByValueAlias = "PinColumnCancellableEventArgsDetail";
    return WebPinColumnCancellableEventArgsDetailDescription;
}(Description));
export { WebPinColumnCancellableEventArgsDetailDescription };
