// igniteui-react-core/esm5/lib/WebPivotUISettingsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebPivotUISettingsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebPivotUISettingsDescription, _super);
    function WebPivotUISettingsDescription() {
        var _this = _super.call(this) || this;
        _this.h = false;
        _this.i = false;
        _this.m = null;
        _this.l = null;
        return _this;
    }
    WebPivotUISettingsDescription.prototype.get_type = function () {
        return "WebPivotUISettings";
    };
    Object.defineProperty(WebPivotUISettingsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotUISettingsDescription.prototype, "showConfiguration", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ShowConfiguration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotUISettingsDescription.prototype, "showRowHeaders", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ShowRowHeaders");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotUISettingsDescription.prototype, "rowLayout", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("RowLayout");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotUISettingsDescription.prototype, "horizontalSummariesPosition", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("HorizontalSummariesPosition");
        },
        enumerable: false,
        configurable: true
    });
    WebPivotUISettingsDescription.$t = markType(WebPivotUISettingsDescription, 'WebPivotUISettingsDescription', Description.$);
    return WebPivotUISettingsDescription;
}(Description));
export { WebPivotUISettingsDescription };
