// igniteui-react-core/esm5/lib/WebProgressBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebProgressBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebProgressBaseDescription, _super);
    function WebProgressBaseDescription() {
        var _this = _super.call(this) || this;
        _this.m = 0;
        _this.n = 0;
        _this.s = null;
        _this.l = 0;
        _this.i = false;
        _this.h = false;
        _this.r = null;
        return _this;
    }
    WebProgressBaseDescription.prototype.get_type = function () {
        return "WebProgressBase";
    };
    Object.defineProperty(WebProgressBaseDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebProgressBaseDescription.prototype, "max", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Max");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebProgressBaseDescription.prototype, "value", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebProgressBaseDescription.prototype, "variant", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Variant");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebProgressBaseDescription.prototype, "animationDuration", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("AnimationDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebProgressBaseDescription.prototype, "indeterminate", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Indeterminate");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebProgressBaseDescription.prototype, "hideLabel", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("HideLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebProgressBaseDescription.prototype, "labelFormat", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("LabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    WebProgressBaseDescription.$t = markType(WebProgressBaseDescription, 'WebProgressBaseDescription', Description.$);
    return WebProgressBaseDescription;
}(Description));
export { WebProgressBaseDescription };
