// igniteui-react-core/esm5/lib/WebPivotDimensionDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebPivotDimensionDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebPivotDimensionDetailDescription, _super);
    function WebPivotDimensionDetailDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.y = null;
        _this.w = null;
        _this.m = false;
        _this.h = null;
        _this.o = false;
        _this.z = null;
        _this.v = null;
        _this.aa = null;
        _this.s = 0;
        _this.n = false;
        _this.x = null;
        return _this;
    }
    WebPivotDimensionDetailDescription.prototype.get_type = function () {
        return "WebPivotDimensionDetail";
    };
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "childLevel", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ChildLevel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "memberName", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("MemberName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "displayName", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("DisplayName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "enabled", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Enabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "filter", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Filter");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "sortable", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Sortable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "sortDirection", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("SortDirection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "dataType", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("DataType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "width", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "level", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Level");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "horizontalSummary", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("HorizontalSummary");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotDimensionDetailDescription.prototype, "memberFunctionRef", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("MemberFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    WebPivotDimensionDetailDescription.$t = markType(WebPivotDimensionDetailDescription, 'WebPivotDimensionDetailDescription', Description.$);
    WebPivotDimensionDetailDescription.__marshalByValue = true;
    WebPivotDimensionDetailDescription.__marshalByValueAlias = "PivotDimensionDetail";
    return WebPivotDimensionDetailDescription;
}(Description));
export { WebPivotDimensionDetailDescription };
