// igniteui-react-core/esm5/lib/WebPaneDragOverEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebPaneDragOverEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebPaneDragOverEventArgsDetailDescription, _super);
    function WebPaneDragOverEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.h = null;
        _this.l = null;
        _this.o = false;
        return _this;
    }
    WebPaneDragOverEventArgsDetailDescription.prototype.get_type = function () {
        return "WebPaneDragOverEventArgsDetail";
    };
    Object.defineProperty(WebPaneDragOverEventArgsDetailDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaneDragOverEventArgsDetailDescription.prototype, "sourcePane", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("SourcePane");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaneDragOverEventArgsDetailDescription.prototype, "panes", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Panes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaneDragOverEventArgsDetailDescription.prototype, "action", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Action");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPaneDragOverEventArgsDetailDescription.prototype, "isValid", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("IsValid");
        },
        enumerable: false,
        configurable: true
    });
    WebPaneDragOverEventArgsDetailDescription.$t = markType(WebPaneDragOverEventArgsDetailDescription, 'WebPaneDragOverEventArgsDetailDescription', Description.$);
    WebPaneDragOverEventArgsDetailDescription.__marshalByValue = true;
    WebPaneDragOverEventArgsDetailDescription.__marshalByValueAlias = "PaneDragOverEventArgsDetail";
    return WebPaneDragOverEventArgsDetailDescription;
}(Description));
export { WebPaneDragOverEventArgsDetailDescription };
