// igniteui-react-core/esm5/lib/WebPivotAggregatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebPivotAggregatorDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebPivotAggregatorDescription, _super);
    function WebPivotAggregatorDescription() {
        var _this = _super.call(this) || this;
        _this.l = null;
        _this.m = null;
        _this.j = null;
        _this.k = null;
        return _this;
    }
    WebPivotAggregatorDescription.prototype.get_type = function () {
        return "WebPivotAggregator";
    };
    Object.defineProperty(WebPivotAggregatorDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotAggregatorDescription.prototype, "key", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Key");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotAggregatorDescription.prototype, "label", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotAggregatorDescription.prototype, "aggregatorName", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("AggregatorName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotAggregatorDescription.prototype, "aggregatorRef", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("AggregatorRef");
        },
        enumerable: false,
        configurable: true
    });
    WebPivotAggregatorDescription.$t = markType(WebPivotAggregatorDescription, 'WebPivotAggregatorDescription', Description.$);
    WebPivotAggregatorDescription.__marshalByValue = true;
    WebPivotAggregatorDescription.__marshalByValueAlias = "PivotAggregator";
    return WebPivotAggregatorDescription;
}(Description));
export { WebPivotAggregatorDescription };
