// igniteui-react-core/esm5/lib/WebPivotGridDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebGridBaseDirectiveDescription } from "./WebGridBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebPivotGridDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebPivotGridDescription, _super);
    function WebPivotGridDescription() {
        var _this = _super.call(this) || this;
        _this.g6 = null;
        _this.g5 = null;
        _this.gp = null;
        _this.gt = false;
        _this.gr = null;
        _this.gv = false;
        _this.gu = false;
        _this.gz = null;
        _this.g3 = null;
        _this.g1 = null;
        _this.g4 = null;
        _this.g0 = null;
        _this.g7 = null;
        _this.g2 = null;
        _this.g8 = null;
        return _this;
    }
    WebPivotGridDescription.prototype.get_type = function () {
        return "WebPivotGrid";
    };
    Object.defineProperty(WebPivotGridDescription.prototype, "valueChipTemplateRef", {
        get: function () {
            return this.g6;
        },
        set: function (a) {
            this.g6 = a;
            this.g("ValueChipTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "rowDimensionHeaderTemplateRef", {
        get: function () {
            return this.g5;
        },
        set: function (a) {
            this.g5 = a;
            this.g("RowDimensionHeaderTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "pivotConfiguration", {
        get: function () {
            return this.gp;
        },
        set: function (a) {
            this.gp = a;
            this.g("PivotConfiguration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "autoGenerateConfig", {
        get: function () {
            return this.gt;
        },
        set: function (a) {
            this.gt = a;
            this.g("AutoGenerateConfig");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "pivotUI", {
        get: function () {
            return this.gr;
        },
        set: function (a) {
            this.gr = a;
            this.g("PivotUI");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "superCompactMode", {
        get: function () {
            return this.gv;
        },
        set: function (a) {
            this.gv = a;
            this.g("SuperCompactMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "defaultExpandState", {
        get: function () {
            return this.gu;
        },
        set: function (a) {
            this.gu = a;
            this.g("DefaultExpandState");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "dataRef", {
        get: function () {
            return this.gz;
        },
        set: function (a) {
            this.gz = a;
            this.g("DataRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "emptyPivotGridTemplateRef", {
        get: function () {
            return this.g3;
        },
        set: function (a) {
            this.g3 = a;
            this.g("EmptyPivotGridTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "dimensionsChangeRef", {
        get: function () {
            return this.g1;
        },
        set: function (a) {
            this.g1 = a;
            this.g("DimensionsChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "pivotConfigurationChangeRef", {
        get: function () {
            return this.g4;
        },
        set: function (a) {
            this.g4 = a;
            this.g("PivotConfigurationChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "dimensionInitRef", {
        get: function () {
            return this.g0;
        },
        set: function (a) {
            this.g0 = a;
            this.g("DimensionInitRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "valueInitRef", {
        get: function () {
            return this.g7;
        },
        set: function (a) {
            this.g7 = a;
            this.g("ValueInitRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "dimensionsSortingExpressionsChangeRef", {
        get: function () {
            return this.g2;
        },
        set: function (a) {
            this.g2 = a;
            this.g("DimensionsSortingExpressionsChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebPivotGridDescription.prototype, "valuesChangeRef", {
        get: function () {
            return this.g8;
        },
        set: function (a) {
            this.g8 = a;
            this.g("ValuesChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    WebPivotGridDescription.$t = markType(WebPivotGridDescription, 'WebPivotGridDescription', WebGridBaseDirectiveDescription.$);
    return WebPivotGridDescription;
}(WebGridBaseDirectiveDescription));
export { WebPivotGridDescription };
