// igniteui-react-core/esm5/lib/WebRowDirectiveDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebRowDirectiveDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebRowDirectiveDescription, _super);
    function WebRowDirectiveDescription() {
        var _this = _super.call(this) || this;
        _this.p = null;
        _this.n = 0;
        _this.h = false;
        _this.j = false;
        _this.i = false;
        return _this;
    }
    WebRowDirectiveDescription.prototype.get_type = function () {
        return "WebRowDirective";
    };
    Object.defineProperty(WebRowDirectiveDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDirectiveDescription.prototype, "data", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("Data");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDirectiveDescription.prototype, "index", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Index");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDirectiveDescription.prototype, "disabled", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDirectiveDescription.prototype, "pinned", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Pinned");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDirectiveDescription.prototype, "expanded", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Expanded");
        },
        enumerable: false,
        configurable: true
    });
    WebRowDirectiveDescription.$t = markType(WebRowDirectiveDescription, 'WebRowDirectiveDescription', Description.$);
    return WebRowDirectiveDescription;
}(Description));
export { WebRowDirectiveDescription };
