// igniteui-react-core/esm5/lib/WebRowDataEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebRowDataEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebRowDataEventArgsDetailDescription, _super);
    function WebRowDataEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.p = null;
        _this.o = null;
        _this.q = null;
        return _this;
    }
    WebRowDataEventArgsDetailDescription.prototype.get_type = function () {
        return "WebRowDataEventArgsDetail";
    };
    Object.defineProperty(WebRowDataEventArgsDetailDescription.prototype, "data", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Data");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataEventArgsDetailDescription.prototype, "rowData", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("RowData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataEventArgsDetailDescription.prototype, "primaryKey", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("PrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataEventArgsDetailDescription.prototype, "rowKey", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("RowKey");
        },
        enumerable: false,
        configurable: true
    });
    WebRowDataEventArgsDetailDescription.$t = markType(WebRowDataEventArgsDetailDescription, 'WebRowDataEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebRowDataEventArgsDetailDescription.__marshalByValue1 = true;
    WebRowDataEventArgsDetailDescription.__marshalByValueAlias1 = "RowDataEventArgsDetail";
    return WebRowDataEventArgsDetailDescription;
}(WebBaseEventArgsDetailDescription));
export { WebRowDataEventArgsDetailDescription };
