// igniteui-react-core/esm5/lib/WebRowDataCancelableEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebRowDataCancelableEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebRowDataCancelableEventArgsDetailDescription, _super);
    function WebRowDataCancelableEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.q = null;
        _this.p = null;
        _this.m = false;
        _this.h = null;
        _this.o = null;
        _this.s = null;
        _this.r = null;
        _this.t = null;
        return _this;
    }
    WebRowDataCancelableEventArgsDetailDescription.prototype.get_type = function () {
        return "WebRowDataCancelableEventArgsDetail";
    };
    Object.defineProperty(WebRowDataCancelableEventArgsDetailDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataCancelableEventArgsDetailDescription.prototype, "cellID", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("CellID");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataCancelableEventArgsDetailDescription.prototype, "oldValue", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("OldValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataCancelableEventArgsDetailDescription.prototype, "newValue", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("NewValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataCancelableEventArgsDetailDescription.prototype, "isAddRow", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("IsAddRow");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataCancelableEventArgsDetailDescription.prototype, "owner", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Owner");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataCancelableEventArgsDetailDescription.prototype, "data", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Data");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataCancelableEventArgsDetailDescription.prototype, "rowData", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("RowData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataCancelableEventArgsDetailDescription.prototype, "primaryKey", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("PrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDataCancelableEventArgsDetailDescription.prototype, "rowKey", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("RowKey");
        },
        enumerable: false,
        configurable: true
    });
    WebRowDataCancelableEventArgsDetailDescription.$t = markType(WebRowDataCancelableEventArgsDetailDescription, 'WebRowDataCancelableEventArgsDetailDescription', Description.$);
    WebRowDataCancelableEventArgsDetailDescription.__marshalByValue = true;
    WebRowDataCancelableEventArgsDetailDescription.__marshalByValueAlias = "RowDataCancelableEventArgsDetail";
    return WebRowDataCancelableEventArgsDetailDescription;
}(Description));
export { WebRowDataCancelableEventArgsDetailDescription };
