// igniteui-react-core/esm5/lib/WebRowDragStartEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebRowDragStartEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebRowDragStartEventArgsDetailDescription, _super);
    function WebRowDragStartEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.m = null;
        _this.h = null;
        _this.k = false;
        _this.n = null;
        return _this;
    }
    WebRowDragStartEventArgsDetailDescription.prototype.get_type = function () {
        return "WebRowDragStartEventArgsDetail";
    };
    Object.defineProperty(WebRowDragStartEventArgsDetailDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDragStartEventArgsDetailDescription.prototype, "dragDirective", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("DragDirective");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDragStartEventArgsDetailDescription.prototype, "dragData", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("DragData");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDragStartEventArgsDetailDescription.prototype, "cancel", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Cancel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowDragStartEventArgsDetailDescription.prototype, "owner", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Owner");
        },
        enumerable: false,
        configurable: true
    });
    WebRowDragStartEventArgsDetailDescription.$t = markType(WebRowDragStartEventArgsDetailDescription, 'WebRowDragStartEventArgsDetailDescription', Description.$);
    WebRowDragStartEventArgsDetailDescription.__marshalByValue = true;
    WebRowDragStartEventArgsDetailDescription.__marshalByValueAlias = "RowDragStartEventArgsDetail";
    return WebRowDragStartEventArgsDetailDescription;
}(Description));
export { WebRowDragStartEventArgsDetailDescription };
