// igniteui-react-core/esm5/lib/WebRangeSliderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebSliderBaseDescription } from "./WebSliderBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebRangeSliderDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebRangeSliderDescription, _super);
    function WebRangeSliderDescription() {
        var _this = _super.call(this) || this;
        _this.aq = 0;
        _this.ar = 0;
        _this.aw = null;
        _this.ax = null;
        _this.av = null;
        _this.au = null;
        return _this;
    }
    WebRangeSliderDescription.prototype.get_type = function () {
        return "WebRangeSlider";
    };
    Object.defineProperty(WebRangeSliderDescription.prototype, "lower", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("Lower");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRangeSliderDescription.prototype, "upper", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("Upper");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRangeSliderDescription.prototype, "thumbLabelLower", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("ThumbLabelLower");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRangeSliderDescription.prototype, "thumbLabelUpper", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("ThumbLabelUpper");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRangeSliderDescription.prototype, "inputRef", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("InputRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRangeSliderDescription.prototype, "changeRef", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    WebRangeSliderDescription.$t = markType(WebRangeSliderDescription, 'WebRangeSliderDescription', WebSliderBaseDescription.$);
    return WebRangeSliderDescription;
}(WebSliderBaseDescription));
export { WebRangeSliderDescription };
