// igniteui-react-core/esm5/lib/WebRatingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebRatingDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebRatingDescription, _super);
    function WebRatingDescription() {
        var _this = _super.call(this) || this;
        _this.t = 0;
        _this.u = 0;
        _this.ad = null;
        _this.af = null;
        _this.v = 0;
        _this.j = false;
        _this.l = false;
        _this.m = false;
        _this.h = false;
        _this.i = false;
        _this.z = null;
        _this.ae = null;
        _this.k = false;
        _this.ab = null;
        _this.ac = null;
        return _this;
    }
    WebRatingDescription.prototype.get_type = function () {
        return "WebRating";
    };
    Object.defineProperty(WebRatingDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "max", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("Max");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "step", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Step");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "label", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("Label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "valueFormat", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("ValueFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "value", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "hoverPreview", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("HoverPreview");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "readOnly", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ReadOnly");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "single", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Single");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "allowReset", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("AllowReset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "disabled", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "defaultValue", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("DefaultValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "name", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "invalid", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Invalid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "changeRef", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRatingDescription.prototype, "hoverRef", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("HoverRef");
        },
        enumerable: false,
        configurable: true
    });
    WebRatingDescription.$t = markType(WebRatingDescription, 'WebRatingDescription', Description.$);
    return WebRatingDescription;
}(Description));
export { WebRatingDescription };
