// igniteui-react-core/esm5/lib/WebRadioGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebRadioGroupDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebRadioGroupDescription, _super);
    function WebRadioGroupDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.j = null;
        _this.k = null;
        _this.l = null;
        _this.i = null;
        return _this;
    }
    WebRadioGroupDescription.prototype.get_type = function () {
        return "WebRadioGroup";
    };
    Object.defineProperty(WebRadioGroupDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioGroupDescription.prototype, "alignment", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Alignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioGroupDescription.prototype, "defaultValue", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("DefaultValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioGroupDescription.prototype, "name", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioGroupDescription.prototype, "value", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioGroupDescription.prototype, "changeRef", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    WebRadioGroupDescription.$t = markType(WebRadioGroupDescription, 'WebRadioGroupDescription', Description.$);
    return WebRadioGroupDescription;
}(Description));
export { WebRadioGroupDescription };
