// igniteui-react-core/esm5/lib/WebRowToggleEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebRowToggleEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebRowToggleEventArgsDetailDescription, _super);
    function WebRowToggleEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.r = null;
        _this.s = null;
        _this.o = false;
        _this.n = false;
        return _this;
    }
    WebRowToggleEventArgsDetailDescription.prototype.get_type = function () {
        return "WebRowToggleEventArgsDetail";
    };
    Object.defineProperty(WebRowToggleEventArgsDetailDescription.prototype, "rowID", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("RowID");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowToggleEventArgsDetailDescription.prototype, "rowKey", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("RowKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowToggleEventArgsDetailDescription.prototype, "expanded", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Expanded");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRowToggleEventArgsDetailDescription.prototype, "cancel", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Cancel");
        },
        enumerable: false,
        configurable: true
    });
    WebRowToggleEventArgsDetailDescription.$t = markType(WebRowToggleEventArgsDetailDescription, 'WebRowToggleEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebRowToggleEventArgsDetailDescription.__marshalByValue1 = true;
    WebRowToggleEventArgsDetailDescription.__marshalByValueAlias1 = "RowToggleEventArgsDetail";
    return WebRowToggleEventArgsDetailDescription;
}(WebBaseEventArgsDetailDescription));
export { WebRowToggleEventArgsDetailDescription };
