// igniteui-react-core/esm5/lib/WebRadioDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebRadioDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebRadioDescription, _super);
    function WebRadioDescription() {
        var _this = _super.call(this) || this;
        _this.w = null;
        _this.h = false;
        _this.u = null;
        _this.j = false;
        _this.l = false;
        _this.i = false;
        _this.v = null;
        _this.k = false;
        _this.s = null;
        _this.t = null;
        _this.r = null;
        return _this;
    }
    WebRadioDescription.prototype.get_type = function () {
        return "WebRadio";
    };
    Object.defineProperty(WebRadioDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "value", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "checked", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Checked");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "labelPosition", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("LabelPosition");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "disabled", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "required", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Required");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "defaultChecked", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("DefaultChecked");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "name", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "invalid", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Invalid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "changeRef", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "focusRef", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("FocusRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebRadioDescription.prototype, "blurRef", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("BlurRef");
        },
        enumerable: false,
        configurable: true
    });
    WebRadioDescription.$t = markType(WebRadioDescription, 'WebRadioDescription', Description.$);
    return WebRadioDescription;
}(Description));
export { WebRadioDescription };
