// igniteui-react-core/esm5/lib/WebSummaryResultDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebSummaryResultDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebSummaryResultDescription, _super);
    function WebSummaryResultDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.o = null;
        _this.k = null;
        _this.i = false;
        return _this;
    }
    WebSummaryResultDescription.prototype.get_type = function () {
        return "WebSummaryResult";
    };
    Object.defineProperty(WebSummaryResultDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSummaryResultDescription.prototype, "key", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Key");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSummaryResultDescription.prototype, "label", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSummaryResultDescription.prototype, "result", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("Result");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSummaryResultDescription.prototype, "defaultFormatting", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("DefaultFormatting");
        },
        enumerable: false,
        configurable: true
    });
    WebSummaryResultDescription.$t = markType(WebSummaryResultDescription, 'WebSummaryResultDescription', Description.$);
    WebSummaryResultDescription.__marshalByValue = true;
    WebSummaryResultDescription.__marshalByValueAlias = "SummaryResult";
    return WebSummaryResultDescription;
}(Description));
export { WebSummaryResultDescription };
