// igniteui-react-core/esm5/lib/WebSelectDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebBaseComboBoxLikeDescription } from "./WebBaseComboBoxLikeDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebSelectDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebSelectDescription, _super);
    function WebSelectDescription() {
        var _this = _super.call(this) || this;
        _this.aq = null;
        _this.t = false;
        _this.q = false;
        _this.aa = 0;
        _this.aj = null;
        _this.an = null;
        _this.ao = null;
        _this.ap = null;
        _this.r = false;
        _this.u = false;
        _this.ac = null;
        _this.ak = null;
        _this.s = false;
        _this.af = null;
        _this.ai = null;
        _this.ae = null;
        _this.am = null;
        _this.al = null;
        _this.ah = null;
        _this.ag = null;
        return _this;
    }
    WebSelectDescription.prototype.get_type = function () {
        return "WebSelect";
    };
    Object.defineProperty(WebSelectDescription.prototype, "value", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "outlined", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("Outlined");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "autofocus", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Autofocus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "distance", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("Distance");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "label", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("Label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "placeholder", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("Placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "placement", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("Placement");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "scrollStrategy", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("ScrollStrategy");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "disabled", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "required", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Required");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "defaultValue", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("DefaultValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "name", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "invalid", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Invalid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "changeRef", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "focusRef", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("FocusRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "blurRef", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("BlurRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "openingRef", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("OpeningRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "openedRef", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("OpenedRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "closingRef", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("ClosingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSelectDescription.prototype, "closedRef", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("ClosedRef");
        },
        enumerable: false,
        configurable: true
    });
    WebSelectDescription.$t = markType(WebSelectDescription, 'WebSelectDescription', WebBaseComboBoxLikeDescription.$);
    return WebSelectDescription;
}(WebBaseComboBoxLikeDescription));
export { WebSelectDescription };
