// igniteui-react-core/esm5/lib/WebSortingEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebSortingEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebSortingEventArgsDetailDescription, _super);
    function WebSortingEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.i = null;
        _this.h = null;
        _this.o = null;
        _this.m = false;
        return _this;
    }
    WebSortingEventArgsDetailDescription.prototype.get_type = function () {
        return "WebSortingEventArgsDetail";
    };
    Object.defineProperty(WebSortingEventArgsDetailDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSortingEventArgsDetailDescription.prototype, "sortingExpressions", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("SortingExpressions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSortingEventArgsDetailDescription.prototype, "groupingExpressions", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("GroupingExpressions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSortingEventArgsDetailDescription.prototype, "owner", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Owner");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSortingEventArgsDetailDescription.prototype, "cancel", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Cancel");
        },
        enumerable: false,
        configurable: true
    });
    WebSortingEventArgsDetailDescription.$t = markType(WebSortingEventArgsDetailDescription, 'WebSortingEventArgsDetailDescription', Description.$);
    WebSortingEventArgsDetailDescription.__marshalByValue = true;
    WebSortingEventArgsDetailDescription.__marshalByValueAlias = "SortingEventArgsDetail";
    return WebSortingEventArgsDetailDescription;
}(Description));
export { WebSortingEventArgsDetailDescription };
