// igniteui-react-core/esm5/lib/WebSliderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebSliderBaseDescription } from "./WebSliderBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebSliderDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebSliderDescription, _super);
    function WebSliderDescription() {
        var _this = _super.call(this) || this;
        _this.as = 0;
        _this.au = null;
        _this.ay = null;
        _this.aq = false;
        _this.ax = null;
        _this.aw = null;
        return _this;
    }
    WebSliderDescription.prototype.get_type = function () {
        return "WebSlider";
    };
    Object.defineProperty(WebSliderDescription.prototype, "value", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderDescription.prototype, "defaultValue", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.g("DefaultValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderDescription.prototype, "name", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderDescription.prototype, "invalid", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("Invalid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderDescription.prototype, "inputRef", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("InputRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderDescription.prototype, "changeRef", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    WebSliderDescription.$t = markType(WebSliderDescription, 'WebSliderDescription', WebSliderBaseDescription.$);
    return WebSliderDescription;
}(WebSliderBaseDescription));
export { WebSliderDescription };
