// igniteui-react-core/esm5/lib/WebSliderBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebSliderBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebSliderBaseDescription, _super);
    function WebSliderBaseDescription() {
        var _this = _super.call(this) || this;
        _this.v = 0;
        _this.u = 0;
        _this.t = 0;
        _this.z = 0;
        _this.j = false;
        _this.k = false;
        _this.n = false;
        _this.y = 0;
        _this.w = 0;
        _this.x = 0;
        _this.aj = null;
        _this.l = false;
        _this.m = false;
        _this.ah = null;
        _this.ak = null;
        _this.ai = null;
        _this.h = null;
        return _this;
    }
    WebSliderBaseDescription.prototype.get_type = function () {
        return "WebSliderBase";
    };
    Object.defineProperty(WebSliderBaseDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "min", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("Min");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "max", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Max");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "lowerBound", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("LowerBound");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "upperBound", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("UpperBound");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "disabled", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "discreteTrack", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("DiscreteTrack");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "hideTooltip", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("HideTooltip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "step", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("Step");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "primaryTicks", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("PrimaryTicks");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "secondaryTicks", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("SecondaryTicks");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "tickOrientation", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("TickOrientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "hidePrimaryLabels", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("HidePrimaryLabels");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "hideSecondaryLabels", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("HideSecondaryLabels");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "locale", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("Locale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "valueFormat", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("ValueFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "tickLabelRotation", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("TickLabelRotation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSliderBaseDescription.prototype, "valueFormatOptions", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("ValueFormatOptions");
        },
        enumerable: false,
        configurable: true
    });
    WebSliderBaseDescription.$t = markType(WebSliderBaseDescription, 'WebSliderBaseDescription', Description.$);
    return WebSliderBaseDescription;
}(Description));
export { WebSliderBaseDescription };
