// igniteui-react-core/esm5/lib/WebSortingExpressionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebBaseEventArgsDescription } from "./WebBaseEventArgsDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebSortingExpressionDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebSortingExpressionDescription, _super);
    function WebSortingExpressionDescription() {
        var _this = _super.call(this) || this;
        _this.t = null;
        _this.s = null;
        _this.p = false;
        _this.m = null;
        return _this;
    }
    WebSortingExpressionDescription.prototype.get_type = function () {
        return "WebSortingExpression";
    };
    Object.defineProperty(WebSortingExpressionDescription.prototype, "fieldName", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("FieldName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSortingExpressionDescription.prototype, "dir", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Dir");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSortingExpressionDescription.prototype, "ignoreCase", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("IgnoreCase");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSortingExpressionDescription.prototype, "strategy", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Strategy");
        },
        enumerable: false,
        configurable: true
    });
    WebSortingExpressionDescription.$t = markType(WebSortingExpressionDescription, 'WebSortingExpressionDescription', WebBaseEventArgsDescription.$);
    WebSortingExpressionDescription.__marshalByValue1 = true;
    WebSortingExpressionDescription.__marshalByValueAlias1 = "SortingExpression";
    return WebSortingExpressionDescription;
}(WebBaseEventArgsDescription));
export { WebSortingExpressionDescription };
