// igniteui-react-core/esm5/lib/WebSplitPaneDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebDockManagerPaneDescription } from "./WebDockManagerPaneDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebSplitPaneDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebSplitPaneDescription, _super);
    function WebSplitPaneDescription() {
        var _this = _super.call(this) || this;
        _this.ae = null;
        _this.ag = null;
        _this.af = null;
        _this.m = null;
        _this.z = 0;
        _this.o = null;
        _this.y = 0;
        _this.x = 0;
        _this.s = false;
        _this.r = false;
        _this.t = false;
        return _this;
    }
    WebSplitPaneDescription.prototype.get_type = function () {
        return "WebSplitPane";
    };
    Object.defineProperty(WebSplitPaneDescription.prototype, "id", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSplitPaneDescription.prototype, "paneType", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("PaneType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSplitPaneDescription.prototype, "orientation", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("Orientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSplitPaneDescription.prototype, "panes", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Panes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSplitPaneDescription.prototype, "size", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("Size");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSplitPaneDescription.prototype, "floatingLocation", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("FloatingLocation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSplitPaneDescription.prototype, "floatingWidth", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("FloatingWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSplitPaneDescription.prototype, "floatingHeight", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("FloatingHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSplitPaneDescription.prototype, "floatingResizable", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("FloatingResizable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSplitPaneDescription.prototype, "allowEmpty", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("AllowEmpty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSplitPaneDescription.prototype, "isMaximized", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("IsMaximized");
        },
        enumerable: false,
        configurable: true
    });
    WebSplitPaneDescription.$t = markType(WebSplitPaneDescription, 'WebSplitPaneDescription', WebDockManagerPaneDescription.$);
    WebSplitPaneDescription.__marshalByValue1 = true;
    WebSplitPaneDescription.__marshalByValueAlias1 = "SplitPane";
    return WebSplitPaneDescription;
}(WebDockManagerPaneDescription));
export { WebSplitPaneDescription };
