// igniteui-react-core/esm5/lib/WebTabGroupPaneDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebDockManagerPaneDescription } from "./WebDockManagerPaneDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebTabGroupPaneDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebTabGroupPaneDescription, _super);
    function WebTabGroupPaneDescription() {
        var _this = _super.call(this) || this;
        _this.y = null;
        _this.z = null;
        _this.m = null;
        _this.u = 0;
        _this.t = 0;
        _this.p = false;
        _this.q = false;
        return _this;
    }
    WebTabGroupPaneDescription.prototype.get_type = function () {
        return "WebTabGroupPane";
    };
    Object.defineProperty(WebTabGroupPaneDescription.prototype, "id", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("Id");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTabGroupPaneDescription.prototype, "paneType", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("PaneType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTabGroupPaneDescription.prototype, "panes", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Panes");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTabGroupPaneDescription.prototype, "size", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Size");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTabGroupPaneDescription.prototype, "selectedIndex", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("SelectedIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTabGroupPaneDescription.prototype, "allowEmpty", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("AllowEmpty");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTabGroupPaneDescription.prototype, "isMaximized", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("IsMaximized");
        },
        enumerable: false,
        configurable: true
    });
    WebTabGroupPaneDescription.$t = markType(WebTabGroupPaneDescription, 'WebTabGroupPaneDescription', WebDockManagerPaneDescription.$);
    WebTabGroupPaneDescription.__marshalByValue1 = true;
    WebTabGroupPaneDescription.__marshalByValueAlias1 = "TabGroupPane";
    return WebTabGroupPaneDescription;
}(WebDockManagerPaneDescription));
export { WebTabGroupPaneDescription };
