// igniteui-react-core/esm5/lib/WebTextareaDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebTextareaDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebTextareaDescription, _super);
    function WebTextareaDescription() {
        var _this = _super.call(this) || this;
        _this.ae = null;
        _this.ad = null;
        _this.aj = null;
        _this.ak = null;
        _this.v = 0;
        _this.w = 0;
        _this.j = false;
        _this.am = null;
        _this.k = false;
        _this.an = null;
        _this.x = 0;
        _this.ao = null;
        _this.m = false;
        _this.ap = null;
        _this.n = false;
        _this.h = false;
        _this.l = false;
        _this.ab = null;
        _this.al = null;
        _this.i = false;
        _this.ai = null;
        _this.ag = null;
        _this.ah = null;
        _this.af = null;
        return _this;
    }
    WebTextareaDescription.prototype.get_type = function () {
        return "WebTextarea";
    };
    Object.defineProperty(WebTextareaDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "autocomplete", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("Autocomplete");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "autocapitalize", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("Autocapitalize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "inputMode", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.g("InputMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "label", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.g("Label");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "maxLength", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("MaxLength");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "minLength", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("MinLength");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "outlined", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("Outlined");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "placeholder", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.g("Placeholder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "readOnly", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("ReadOnly");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "resize", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.g("Resize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "rows", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("Rows");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "value", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "spellcheck", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Spellcheck");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "wrap", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.g("Wrap");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "validateOnly", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("ValidateOnly");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "disabled", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("Disabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "required", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("Required");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "defaultValue", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("DefaultValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "name", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.g("Name");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "invalid", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Invalid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "inputRef", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.g("InputRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "changeRef", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("ChangeRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "focusRef", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.g("FocusRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebTextareaDescription.prototype, "blurRef", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("BlurRef");
        },
        enumerable: false,
        configurable: true
    });
    WebTextareaDescription.$t = markType(WebTextareaDescription, 'WebTextareaDescription', Description.$);
    return WebTextareaDescription;
}(Description));
export { WebTextareaDescription };
