// igniteui-react-core/esm5/lib/XYChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DomainChartDescription } from "./DomainChartDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var XYChartDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(XYChartDescription, _super);
    function XYChartDescription() {
        var _this = _super.call(this) || this;
        _this.t5 = null;
        _this.un = null;
        _this.sa = 0;
        _this.sc = 0;
        _this.sb = 0;
        _this.r9 = 0;
        _this.sr = 0;
        _this.st = 0;
        _this.ss = 0;
        _this.sq = 0;
        _this.ua = null;
        _this.us = null;
        _this.t3 = null;
        _this.t4 = null;
        _this.sl = 0;
        _this.s2 = 0;
        _this.sk = 0;
        _this.s1 = 0;
        _this.sn = 0;
        _this.s4 = 0;
        _this.sm = 0;
        _this.s3 = 0;
        _this.sj = 0;
        _this.s0 = 0;
        _this.ul = null;
        _this.u3 = null;
        _this.ub = null;
        _this.ut = null;
        _this.um = null;
        _this.u4 = null;
        _this.t9 = null;
        _this.ur = null;
        _this.ue = null;
        _this.uw = null;
        _this.sd = 0;
        _this.su = 0;
        _this.se = 0;
        _this.sv = 0;
        _this.ug = null;
        _this.uy = null;
        _this.uh = null;
        _this.uz = null;
        _this.sf = 0;
        _this.sw = 0;
        _this.sg = 0;
        _this.sx = 0;
        _this.ui = null;
        _this.u0 = null;
        _this.sh = 0;
        _this.sy = 0;
        _this.uj = null;
        _this.u1 = null;
        _this.uf = null;
        _this.ux = null;
        _this.r8 = 0;
        _this.sp = 0;
        _this.r7 = 0;
        _this.so = 0;
        _this.si = 0;
        _this.sz = 0;
        _this.r3 = false;
        _this.r4 = false;
        _this.uk = null;
        _this.u2 = null;
        _this.t7 = null;
        _this.up = null;
        _this.uc = null;
        _this.uu = null;
        _this.ud = null;
        _this.uv = null;
        _this.uq = null;
        _this.t8 = null;
        _this.t6 = null;
        _this.rz = null;
        _this.uo = null;
        _this.r0 = null;
        return _this;
    }
    XYChartDescription.prototype.get_type = function () {
        return "XYChart";
    };
    Object.defineProperty(XYChartDescription.prototype, "xAxisFormatLabelRef", {
        get: function () {
            return this.t5;
        },
        set: function (a) {
            this.t5 = a;
            this.g("XAxisFormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisFormatLabelRef", {
        get: function () {
            return this.un;
        },
        set: function (a) {
            this.un = a;
            this.g("YAxisFormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelLeftMargin", {
        get: function () {
            return this.sa;
        },
        set: function (a) {
            this.sa = a;
            this.g("XAxisLabelLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelTopMargin", {
        get: function () {
            return this.sc;
        },
        set: function (a) {
            this.sc = a;
            this.g("XAxisLabelTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelRightMargin", {
        get: function () {
            return this.sb;
        },
        set: function (a) {
            this.sb = a;
            this.g("XAxisLabelRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelBottomMargin", {
        get: function () {
            return this.r9;
        },
        set: function (a) {
            this.r9 = a;
            this.g("XAxisLabelBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelLeftMargin", {
        get: function () {
            return this.sr;
        },
        set: function (a) {
            this.sr = a;
            this.g("YAxisLabelLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelTopMargin", {
        get: function () {
            return this.st;
        },
        set: function (a) {
            this.st = a;
            this.g("YAxisLabelTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelRightMargin", {
        get: function () {
            return this.ss;
        },
        set: function (a) {
            this.ss = a;
            this.g("YAxisLabelRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelBottomMargin", {
        get: function () {
            return this.sq;
        },
        set: function (a) {
            this.sq = a;
            this.g("YAxisLabelBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelTextColor", {
        get: function () {
            return this.ua;
        },
        set: function (a) {
            this.ua = a;
            this.g("XAxisLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelTextColor", {
        get: function () {
            return this.us;
        },
        set: function (a) {
            this.us = a;
            this.g("YAxisLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "actualXAxisLabelTextColor", {
        get: function () {
            return this.t3;
        },
        set: function (a) {
            this.t3 = a;
            this.g("ActualXAxisLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "actualYAxisLabelTextColor", {
        get: function () {
            return this.t4;
        },
        set: function (a) {
            this.t4 = a;
            this.g("ActualYAxisLabelTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTitleMargin", {
        get: function () {
            return this.sl;
        },
        set: function (a) {
            this.sl = a;
            this.g("XAxisTitleMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTitleMargin", {
        get: function () {
            return this.s2;
        },
        set: function (a) {
            this.s2 = a;
            this.g("YAxisTitleMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTitleLeftMargin", {
        get: function () {
            return this.sk;
        },
        set: function (a) {
            this.sk = a;
            this.g("XAxisTitleLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTitleLeftMargin", {
        get: function () {
            return this.s1;
        },
        set: function (a) {
            this.s1 = a;
            this.g("YAxisTitleLeftMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTitleTopMargin", {
        get: function () {
            return this.sn;
        },
        set: function (a) {
            this.sn = a;
            this.g("XAxisTitleTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTitleTopMargin", {
        get: function () {
            return this.s4;
        },
        set: function (a) {
            this.s4 = a;
            this.g("YAxisTitleTopMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTitleRightMargin", {
        get: function () {
            return this.sm;
        },
        set: function (a) {
            this.sm = a;
            this.g("XAxisTitleRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTitleRightMargin", {
        get: function () {
            return this.s3;
        },
        set: function (a) {
            this.s3 = a;
            this.g("YAxisTitleRightMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTitleBottomMargin", {
        get: function () {
            return this.sj;
        },
        set: function (a) {
            this.sj = a;
            this.g("XAxisTitleBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTitleBottomMargin", {
        get: function () {
            return this.s0;
        },
        set: function (a) {
            this.s0 = a;
            this.g("YAxisTitleBottomMargin");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTitleTextColor", {
        get: function () {
            return this.ul;
        },
        set: function (a) {
            this.ul = a;
            this.g("XAxisTitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTitleTextColor", {
        get: function () {
            return this.u3;
        },
        set: function (a) {
            this.u3 = a;
            this.g("YAxisTitleTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelTextStyle", {
        get: function () {
            return this.ub;
        },
        set: function (a) {
            this.ub = a;
            this.g("XAxisLabelTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelTextStyle", {
        get: function () {
            return this.ut;
        },
        set: function (a) {
            this.ut = a;
            this.g("YAxisLabelTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTitleTextStyle", {
        get: function () {
            return this.um;
        },
        set: function (a) {
            this.um = a;
            this.g("XAxisTitleTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTitleTextStyle", {
        get: function () {
            return this.u4;
        },
        set: function (a) {
            this.u4 = a;
            this.g("YAxisTitleTextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelRef", {
        get: function () {
            return this.t9;
        },
        set: function (a) {
            this.t9 = a;
            this.g("XAxisLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelRef", {
        get: function () {
            return this.ur;
        },
        set: function (a) {
            this.ur = a;
            this.g("YAxisLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisMajorStroke", {
        get: function () {
            return this.ue;
        },
        set: function (a) {
            this.ue = a;
            this.g("XAxisMajorStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisMajorStroke", {
        get: function () {
            return this.uw;
        },
        set: function (a) {
            this.uw = a;
            this.g("YAxisMajorStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisMajorStrokeThickness", {
        get: function () {
            return this.sd;
        },
        set: function (a) {
            this.sd = a;
            this.g("XAxisMajorStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisMajorStrokeThickness", {
        get: function () {
            return this.su;
        },
        set: function (a) {
            this.su = a;
            this.g("YAxisMajorStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisMinorStrokeThickness", {
        get: function () {
            return this.se;
        },
        set: function (a) {
            this.se = a;
            this.g("XAxisMinorStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisMinorStrokeThickness", {
        get: function () {
            return this.sv;
        },
        set: function (a) {
            this.sv = a;
            this.g("YAxisMinorStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisStrip", {
        get: function () {
            return this.ug;
        },
        set: function (a) {
            this.ug = a;
            this.g("XAxisStrip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisStrip", {
        get: function () {
            return this.uy;
        },
        set: function (a) {
            this.uy = a;
            this.g("YAxisStrip");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisStroke", {
        get: function () {
            return this.uh;
        },
        set: function (a) {
            this.uh = a;
            this.g("XAxisStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisStroke", {
        get: function () {
            return this.uz;
        },
        set: function (a) {
            this.uz = a;
            this.g("YAxisStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisStrokeThickness", {
        get: function () {
            return this.sf;
        },
        set: function (a) {
            this.sf = a;
            this.g("XAxisStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisStrokeThickness", {
        get: function () {
            return this.sw;
        },
        set: function (a) {
            this.sw = a;
            this.g("YAxisStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTickLength", {
        get: function () {
            return this.sg;
        },
        set: function (a) {
            this.sg = a;
            this.g("XAxisTickLength");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTickLength", {
        get: function () {
            return this.sx;
        },
        set: function (a) {
            this.sx = a;
            this.g("YAxisTickLength");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTickStroke", {
        get: function () {
            return this.ui;
        },
        set: function (a) {
            this.ui = a;
            this.g("XAxisTickStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTickStroke", {
        get: function () {
            return this.u0;
        },
        set: function (a) {
            this.u0 = a;
            this.g("YAxisTickStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTickStrokeThickness", {
        get: function () {
            return this.sh;
        },
        set: function (a) {
            this.sh = a;
            this.g("XAxisTickStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTickStrokeThickness", {
        get: function () {
            return this.sy;
        },
        set: function (a) {
            this.sy = a;
            this.g("YAxisTickStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTitle", {
        get: function () {
            return this.uj;
        },
        set: function (a) {
            this.uj = a;
            this.g("XAxisTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTitle", {
        get: function () {
            return this.u1;
        },
        set: function (a) {
            this.u1 = a;
            this.g("YAxisTitle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisMinorStroke", {
        get: function () {
            return this.uf;
        },
        set: function (a) {
            this.uf = a;
            this.g("XAxisMinorStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisMinorStroke", {
        get: function () {
            return this.ux;
        },
        set: function (a) {
            this.ux = a;
            this.g("YAxisMinorStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelAngle", {
        get: function () {
            return this.r8;
        },
        set: function (a) {
            this.r8 = a;
            this.g("XAxisLabelAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelAngle", {
        get: function () {
            return this.sp;
        },
        set: function (a) {
            this.sp = a;
            this.g("YAxisLabelAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisExtent", {
        get: function () {
            return this.r7;
        },
        set: function (a) {
            this.r7 = a;
            this.g("XAxisExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisExtent", {
        get: function () {
            return this.so;
        },
        set: function (a) {
            this.so = a;
            this.g("YAxisExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTitleAngle", {
        get: function () {
            return this.si;
        },
        set: function (a) {
            this.si = a;
            this.g("XAxisTitleAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTitleAngle", {
        get: function () {
            return this.sz;
        },
        set: function (a) {
            this.sz = a;
            this.g("YAxisTitleAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisInverted", {
        get: function () {
            return this.r3;
        },
        set: function (a) {
            this.r3 = a;
            this.g("XAxisInverted");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisInverted", {
        get: function () {
            return this.r4;
        },
        set: function (a) {
            this.r4 = a;
            this.g("YAxisInverted");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisTitleAlignment", {
        get: function () {
            return this.uk;
        },
        set: function (a) {
            this.uk = a;
            this.g("XAxisTitleAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisTitleAlignment", {
        get: function () {
            return this.u2;
        },
        set: function (a) {
            this.u2 = a;
            this.g("YAxisTitleAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelHorizontalAlignment", {
        get: function () {
            return this.t7;
        },
        set: function (a) {
            this.t7 = a;
            this.g("XAxisLabelHorizontalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelHorizontalAlignment", {
        get: function () {
            return this.up;
        },
        set: function (a) {
            this.up = a;
            this.g("YAxisLabelHorizontalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelVerticalAlignment", {
        get: function () {
            return this.uc;
        },
        set: function (a) {
            this.uc = a;
            this.g("XAxisLabelVerticalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelVerticalAlignment", {
        get: function () {
            return this.uu;
        },
        set: function (a) {
            this.uu = a;
            this.g("YAxisLabelVerticalAlignment");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelVisibility", {
        get: function () {
            return this.ud;
        },
        set: function (a) {
            this.ud = a;
            this.g("XAxisLabelVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelVisibility", {
        get: function () {
            return this.uv;
        },
        set: function (a) {
            this.uv = a;
            this.g("YAxisLabelVisibility");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelLocation", {
        get: function () {
            return this.uq;
        },
        set: function (a) {
            this.uq = a;
            this.g("YAxisLabelLocation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelLocation", {
        get: function () {
            return this.t8;
        },
        set: function (a) {
            this.t8 = a;
            this.g("XAxisLabelLocation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelFormat", {
        get: function () {
            return this.t6;
        },
        set: function (a) {
            this.t6 = a;
            this.g("XAxisLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "xAxisLabelFormatSpecifiers", {
        get: function () {
            return this.rz;
        },
        set: function (a) {
            this.rz = a;
            this.g("XAxisLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelFormat", {
        get: function () {
            return this.uo;
        },
        set: function (a) {
            this.uo = a;
            this.g("YAxisLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChartDescription.prototype, "yAxisLabelFormatSpecifiers", {
        get: function () {
            return this.r0;
        },
        set: function (a) {
            this.r0 = a;
            this.g("YAxisLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    XYChartDescription.$t = markType(XYChartDescription, 'XYChartDescription', DomainChartDescription.$);
    return XYChartDescription;
}(DomainChartDescription));
export { XYChartDescription };
