// igniteui-react-core/esm5/lib/ZoomSliderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ZoomSliderDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ZoomSliderDescription, _super);
    function ZoomSliderDescription() {
        var _this = _super.call(this) || this;
        _this.bx = null;
        _this.a8 = null;
        _this.a5 = null;
        _this.a3 = 0;
        _this.z = 0;
        _this.ab = 0;
        _this.l = 0;
        _this.h = null;
        _this.aa = 0;
        _this.ad = 0;
        _this.o = 0;
        _this.ag = 0;
        _this.af = 0;
        _this.m = 0;
        _this.bp = null;
        _this.bm = null;
        _this.x = 0;
        _this.s = 0;
        _this.be = null;
        _this.bn = null;
        _this.bf = null;
        _this.bo = null;
        _this.bg = null;
        _this.y = 0;
        _this.t = 0;
        _this.w = 0;
        _this.r = 0;
        _this.bk = null;
        _this.bl = null;
        _this.v = 0;
        _this.bc = null;
        _this.bd = null;
        _this.q = 0;
        _this.a6 = null;
        _this.a7 = null;
        _this.n = 0;
        _this.bq = null;
        _this.br = null;
        _this.ac = 0;
        _this.bs = null;
        _this.bh = null;
        _this.bj = null;
        _this.bi = null;
        _this.u = 0;
        _this.a9 = null;
        _this.bb = null;
        _this.ba = null;
        _this.p = 0;
        _this.j = false;
        _this.bu = null;
        _this.ae = 0;
        _this.bv = null;
        _this.bw = null;
        _this.bt = null;
        _this.by = null;
        return _this;
    }
    Object.defineProperty(ZoomSliderDescription.prototype, "width", {
        get: function () {
            return this.bx;
        },
        set: function (a) {
            this.bx = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "height", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "background", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    ZoomSliderDescription.prototype.get_type = function () {
        return "ZoomSlider";
    };
    Object.defineProperty(ZoomSliderDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "panTransitionDuration", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.g("PanTransitionDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "maxZoomWidth", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.g("MaxZoomWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.g("PixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "actualPixelScalingRatio", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("ActualPixelScalingRatio");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "windowRect", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.g("WindowRect");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "minZoomWidth", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.g("MinZoomWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "startInset", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.g("StartInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "endInset", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("EndInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "trackStartInset", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.g("TrackStartInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "trackEndInset", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.g("TrackEndInset");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "barExtent", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("BarExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "orientation", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.g("Orientation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerThumbBrush", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            this.bm = a;
            this.g("LowerThumbBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerThumbStrokeThickness", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("LowerThumbStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherThumbStrokeThickness", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("HigherThumbStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherThumbBrush", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.g("HigherThumbBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerThumbOutline", {
        get: function () {
            return this.bn;
        },
        set: function (a) {
            this.bn = a;
            this.g("LowerThumbOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherThumbOutline", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.g("HigherThumbOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerThumbRidgesBrush", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            this.bo = a;
            this.g("LowerThumbRidgesBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherThumbRidgesBrush", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.g("HigherThumbRidgesBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerThumbWidth", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("LowerThumbWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherThumbWidth", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("HigherThumbWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerThumbHeight", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("LowerThumbHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherThumbHeight", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("HigherThumbHeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerShadeBrush", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.g("LowerShadeBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerShadeOutline", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.g("LowerShadeOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerShadeStrokeThickness", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("LowerShadeStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherShadeBrush", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.g("HigherShadeBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherShadeOutline", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.g("HigherShadeOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherShadeStrokeThickness", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("HigherShadeStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "barBrush", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.g("BarBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "barOutline", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.g("BarOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "barStrokeThickness", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("BarStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "rangeThumbBrush", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            this.bq = a;
            this.g("RangeThumbBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "rangeThumbOutline", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.g("RangeThumbOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "rangeThumbStrokeThickness", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.g("RangeThumbStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "rangeThumbRidgesBrush", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.g("RangeThumbRidgesBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerCalloutBrush", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.g("LowerCalloutBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerCalloutTextColor", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.g("LowerCalloutTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerCalloutOutline", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.g("LowerCalloutOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "lowerCalloutStrokeThickness", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("LowerCalloutStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherCalloutBrush", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.g("HigherCalloutBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherCalloutTextColor", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.g("HigherCalloutTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherCalloutOutline", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.g("HigherCalloutOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "higherCalloutStrokeThickness", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("HigherCalloutStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "areThumbCalloutsEnabled", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("AreThumbCalloutsEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "thumbCalloutFontFamily", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.g("ThumbCalloutFontFamily");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "thumbCalloutFontSize", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.g("ThumbCalloutFontSize");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "thumbCalloutFontStyle", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.g("ThumbCalloutFontStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "thumbCalloutFontWeight", {
        get: function () {
            return this.bw;
        },
        set: function (a) {
            this.bw = a;
            this.g("ThumbCalloutFontWeight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "resolvingAxisValueRef", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.g("ResolvingAxisValueRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ZoomSliderDescription.prototype, "windowRectChangedRef", {
        get: function () {
            return this.by;
        },
        set: function (a) {
            this.by = a;
            this.g("WindowRectChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    ZoomSliderDescription.$t = markType(ZoomSliderDescription, 'ZoomSliderDescription', Description.$);
    return ZoomSliderDescription;
}(Description));
export { ZoomSliderDescription };
