// igniteui-react-charts/esm2015/lib/AbsoluteVolumeOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { AbsoluteVolumeOscillatorIndicatorStrategy } from "./AbsoluteVolumeOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let AbsoluteVolumeOscillatorIndicator = /*@__PURE__*/ (() => {
    class AbsoluteVolumeOscillatorIndicator extends StrategyBasedIndicator {
        bj() {
            return new AbsoluteVolumeOscillatorIndicator();
        }
        get_abv() {
            return new AbsoluteVolumeOscillatorIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return AbsoluteVolumeOscillatorIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get shortPeriod() {
            return typeGetValue(this.c(AbsoluteVolumeOscillatorIndicator.ach));
        }
        set shortPeriod(a) {
            this.h(AbsoluteVolumeOscillatorIndicator.ach, a);
        }
        ab3() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(AbsoluteVolumeOscillatorIndicator.acg));
        }
        set longPeriod(a) {
            this.h(AbsoluteVolumeOscillatorIndicator.acg, a);
        }
        ab0() {
            return this.longPeriod;
        }
    }
    AbsoluteVolumeOscillatorIndicator.$t = /*@__PURE__*/ markType(AbsoluteVolumeOscillatorIndicator, 'AbsoluteVolumeOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    AbsoluteVolumeOscillatorIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acf(10, AbsoluteVolumeOscillatorIndicator.$);
    AbsoluteVolumeOscillatorIndicator.acg = /*@__PURE__*/ StrategyBasedIndicator.acb(30, AbsoluteVolumeOscillatorIndicator.$);
    return AbsoluteVolumeOscillatorIndicator;
})();
