// igniteui-react-charts/esm2015/lib/AnnotationAppearanceMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available behaviors for how annotations are distinguished from series.
 */
export var AnnotationAppearanceMode = /*@__PURE__*/ (function (AnnotationAppearanceMode) {
    /**
     * Let the annotation layer decide what appearance to use.
     */
    AnnotationAppearanceMode[AnnotationAppearanceMode["Auto"] = 0] = "Auto";
    /**
     * Take the target series color and shift the brightness.
     */
    AnnotationAppearanceMode[AnnotationAppearanceMode["BrightnessShift"] = 1] = "BrightnessShift";
    /**
     * Take the target series color and shift the saturation.
     */
    AnnotationAppearanceMode[AnnotationAppearanceMode["SaturationShift"] = 2] = "SaturationShift";
    /**
     * Take the target series color and apply a dash pattern.
     */
    AnnotationAppearanceMode[AnnotationAppearanceMode["DashPattern"] = 3] = "DashPattern";
    /**
     * Take the target series color and shift the opacity.
     */
    AnnotationAppearanceMode[AnnotationAppearanceMode["OpacityShift"] = 4] = "OpacityShift";
    return AnnotationAppearanceMode;
})({});
/**
 * @hidden
 */
export let AnnotationAppearanceMode_$type = /*@__PURE__*/ markEnum('AnnotationAppearanceMode', 'Auto,0|BrightnessShift,1|SaturationShift,2|DashPattern,3|OpacityShift,4');
