// igniteui-react-charts/esm2015/lib/AssigningSeriesShapeStyleEventArgsBase.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AssigningSeriesStyleEventArgsBase } from "./AssigningSeriesStyleEventArgsBase";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let AssigningSeriesShapeStyleEventArgsBase = /*@__PURE__*/ (() => {
    class AssigningSeriesShapeStyleEventArgsBase extends AssigningSeriesStyleEventArgsBase {
        constructor() {
            super(...arguments);
            this.ae = 0;
            this.af = null;
            this.ag = 0;
            this.ac = 0;
            this.ad = 0;
        }
    }
    AssigningSeriesShapeStyleEventArgsBase.$t = /*@__PURE__*/ markType(AssigningSeriesShapeStyleEventArgsBase, 'AssigningSeriesShapeStyleEventArgsBase', AssigningSeriesStyleEventArgsBase.$);
    return AssigningSeriesShapeStyleEventArgsBase;
})();
