// igniteui-react-charts/esm2015/lib/AverageTrueRangeIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AverageTrueRangeIndicatorStrategy } from "./AverageTrueRangeIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let AverageTrueRangeIndicator = /*@__PURE__*/ (() => {
    class AverageTrueRangeIndicator extends StrategyBasedIndicator {
        bj() {
            return new AverageTrueRangeIndicator();
        }
        get_abv() {
            return new AverageTrueRangeIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return AverageTrueRangeIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(AverageTrueRangeIndicator.ach));
        }
        set acg(a) {
            this.h(AverageTrueRangeIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    AverageTrueRangeIndicator.$t = /*@__PURE__*/ markType(AverageTrueRangeIndicator, 'AverageTrueRangeIndicator', StrategyBasedIndicator.$);
    AverageTrueRangeIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(14, AverageTrueRangeIndicator.$);
    return AverageTrueRangeIndicator;
})();
