// igniteui-react-charts/esm2015/lib/AverageDirectionalIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AverageDirectionalIndexIndicatorStrategy } from "./AverageDirectionalIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let AverageDirectionalIndexIndicator = /*@__PURE__*/ (() => {
    class AverageDirectionalIndexIndicator extends StrategyBasedIndicator {
        bj() {
            return new AverageDirectionalIndexIndicator();
        }
        get_abv() {
            return new AverageDirectionalIndexIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return AverageDirectionalIndexIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(AverageDirectionalIndexIndicator.ach));
        }
        set acg(a) {
            this.h(AverageDirectionalIndexIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    AverageDirectionalIndexIndicator.$t = /*@__PURE__*/ markType(AverageDirectionalIndexIndicator, 'AverageDirectionalIndexIndicator', StrategyBasedIndicator.$);
    AverageDirectionalIndexIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(14, AverageDirectionalIndexIndicator.$);
    return AverageDirectionalIndexIndicator;
})();
