// igniteui-react-charts/esm2015/lib/AxisOrientation.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available orientations for an axis.
 */
export var AxisOrientation = /*@__PURE__*/ (function (AxisOrientation) {
    /**
     * Specifies a horizontal axis.
     */
    AxisOrientation[AxisOrientation["Horizontal"] = 0] = "Horizontal";
    /**
     * Specifies a vertical axis.
     */
    AxisOrientation[AxisOrientation["Vertical"] = 1] = "Vertical";
    /**
     * Specifies an angular axis used by the polar series.
     */
    AxisOrientation[AxisOrientation["Angular"] = 2] = "Angular";
    /**
     * Specifies a radial axis used by radial series.
     */
    AxisOrientation[AxisOrientation["Radial"] = 3] = "Radial";
    return AxisOrientation;
})({});
/**
 * @hidden
 */
export let AxisOrientation_$type = /*@__PURE__*/ markEnum('AxisOrientation', 'Horizontal,0|Vertical,1|Angular,2|Radial,3');
