// igniteui-react-charts/esm2015/lib/AxisLabelsLocation.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available locations of the axis labels in the chart.
 */
export var AxisLabelsLocation = /*@__PURE__*/ (function (AxisLabelsLocation) {
    /**
     * Places the axis labels at the top, outside of the plotting area.
     */
    AxisLabelsLocation[AxisLabelsLocation["OutsideTop"] = 0] = "OutsideTop";
    /**
     * Places the axis labels at the bottom, outside of the plotting area
     */
    AxisLabelsLocation[AxisLabelsLocation["OutsideBottom"] = 1] = "OutsideBottom";
    /**
     * Places the axis labels to the left, outside of the plotting area.
     */
    AxisLabelsLocation[AxisLabelsLocation["OutsideLeft"] = 2] = "OutsideLeft";
    /**
     * Places the axis labels to the right, outside of the plotting area.
     */
    AxisLabelsLocation[AxisLabelsLocation["OutsideRight"] = 3] = "OutsideRight";
    /**
     * Places the axis labels inside the plotting area above the axis line.
     * CrossingAxis should be set for this setting to take effect.
     */
    AxisLabelsLocation[AxisLabelsLocation["InsideTop"] = 4] = "InsideTop";
    /**
     * Places the axis labels inside the plotting area below the axis line.
     * CrossingAxis should be set for this setting to take effect.
     */
    AxisLabelsLocation[AxisLabelsLocation["InsideBottom"] = 5] = "InsideBottom";
    /**
     * Places the axis labels inside the plotting area and to the left of the axis line.
     * CrossingAxis should be set for this setting to take effect.
     */
    AxisLabelsLocation[AxisLabelsLocation["InsideLeft"] = 6] = "InsideLeft";
    /**
     * Places the axis labels inside the plotting area and to the right of the axis line.
     * CrossingAxis should be set for this setting to take effect.
     */
    AxisLabelsLocation[AxisLabelsLocation["InsideRight"] = 7] = "InsideRight";
    return AxisLabelsLocation;
})({});
/**
 * @hidden
 */
export let AxisLabelsLocation_$type = /*@__PURE__*/ markEnum('AxisLabelsLocation', 'OutsideTop,0|OutsideBottom,1|OutsideLeft,2|OutsideRight,3|InsideTop,4|InsideBottom,5|InsideLeft,6|InsideRight,7');
