// igniteui-react-charts/esm2015/lib/AxisRenderingParametersBase.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { RangeInfo } from "./RangeInfo";
/**
 * @hidden
 */
export let AxisRenderingParametersBase = /*@__PURE__*/ (() => {
    class AxisRenderingParametersBase extends Base {
        constructor() {
            super();
            this.o = 0;
            this.aa = null;
            this.ab = null;
            this.z = null;
            this.y = null;
            this.k = null;
            this.e = null;
            this.f = null;
            this.x = null;
            this.v = null;
            this.w = null;
            this.u = null;
            this.m = 0;
            this.l = 0;
            this.n = 0;
            this.r = 0;
            this.t = null;
            this.p = 0;
            this.g = false;
            this.h = false;
            this.j = false;
            this.b = null;
            this.a = null;
            this.c = null;
            this.d = null;
            this.q = 0;
            this.s = 0;
            this.s = -1;
            this.k = new List$1(RangeInfo.$, 0);
        }
        i() {
            return false;
        }
    }
    AxisRenderingParametersBase.$t = /*@__PURE__*/ markType(AxisRenderingParametersBase, 'AxisRenderingParametersBase');
    return AxisRenderingParametersBase;
})();
