// igniteui-react-charts/esm2015/lib/AxisTitlePosition.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available positions for axis title in respect to axis labels
 */
export var AxisTitlePosition = /*@__PURE__*/ (function (AxisTitlePosition) {
    /**
     * Displays the axis title in automatically selected position based on orientation and location of the axis:
     * <para/> - left title position for vertical axis located on left of the chart,
     * <para/> - right title position for vertical axis located on right of the chart,
     * <para/> - top title position for horizontal axis located on top of the chart,
     * <para/> - bottom title position for horizontal axis located on bottom of the chart,
     */
    AxisTitlePosition[AxisTitlePosition["Auto"] = 0] = "Auto";
    /**
     * Displays the axis title to the left of labels on vertical axis
     */
    AxisTitlePosition[AxisTitlePosition["Left"] = 1] = "Left";
    /**
     * Displays the axis title to the right of labels on vertical axis
     */
    AxisTitlePosition[AxisTitlePosition["Right"] = 2] = "Right";
    /**
     * Displays the axis title on top of labels on horizontal axis
     */
    AxisTitlePosition[AxisTitlePosition["Top"] = 3] = "Top";
    /**
     * Displays the axis title on bottom of labels on horizontal axis
     */
    AxisTitlePosition[AxisTitlePosition["Bottom"] = 4] = "Bottom";
    return AxisTitlePosition;
})({});
/**
 * @hidden
 */
export let AxisTitlePosition_$type = /*@__PURE__*/ markEnum('AxisTitlePosition', 'Auto,0|Left,1|Right,2|Top,3|Bottom,4');
