// igniteui-react-charts/esm2015/lib/AxisRange.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, IEquatable$1_$type, typeCast, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let AxisRange = /*@__PURE__*/ (() => {
    class AxisRange extends Base {
        constructor(a, b) {
            super();
            this.c = 0;
            this.b = 0;
            this.c = a;
            this.b = b;
        }
        equals(a) {
            return a != null && this.c == a.c && this.b == a.b;
        }
        equals1(a) {
            return this.equals(typeCast(AxisRange.$, a));
        }
        getHashCode() {
            return super.getHashCode();
        }
    }
    AxisRange.$t = /*@__PURE__*/ /*@__PURE__*/ markType(AxisRange, 'AxisRange', Base.$, [/*@__PURE__*/ IEquatable$1_$type.specialize(-1)]).initSelfReferences();
    return AxisRange;
})();
