// igniteui-react-charts/esm2015/lib/AxisRangeBufferMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available locations of the axis labels in the chart.
 */
export var AxisRangeBufferMode = /*@__PURE__*/ (function (AxisRangeBufferMode) {
    /**
     * The range buffer mode of the axis is decided automatically.
     */
    AxisRangeBufferMode[AxisRangeBufferMode["Auto"] = 0] = "Auto";
    /**
     * The range buffer of the axis is defined by the additional requirements of the series to avoid truncating visuals, including visible range margins.
     */
    AxisRangeBufferMode[AxisRangeBufferMode["Series"] = 1] = "Series";
    /**
     * The range minimum buffer of the axis is defined by the additional requirements of the series to avoid truncating visuals, including visible range margins.
     */
    AxisRangeBufferMode[AxisRangeBufferMode["SeriesMinimum"] = 2] = "SeriesMinimum";
    /**
     * The range maxium buffer of the axis is defined by the additional requirements of the series to avoid truncating visuals, including visible range margins.
     */
    AxisRangeBufferMode[AxisRangeBufferMode["SeriesMaximum"] = 3] = "SeriesMaximum";
    /**
     * No extra range buffer should be added.
     */
    AxisRangeBufferMode[AxisRangeBufferMode["None"] = 4] = "None";
    return AxisRangeBufferMode;
})({});
/**
 * @hidden
 */
export let AxisRangeBufferMode_$type = /*@__PURE__*/ markEnum('AxisRangeBufferMode', 'Auto,0|Series,1|SeriesMinimum,2|SeriesMaximum,3|None,4');
