// igniteui-react-charts/esm2015/lib/BollingerBandWidthIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { BollingerBandWidthIndicatorStrategy } from "./BollingerBandWidthIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let BollingerBandWidthIndicator = /*@__PURE__*/ (() => {
    class BollingerBandWidthIndicator extends StrategyBasedIndicator {
        bj() {
            return new BollingerBandWidthIndicator();
        }
        get_abv() {
            return new BollingerBandWidthIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return BollingerBandWidthIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get ach() {
            return typeGetValue(this.c(BollingerBandWidthIndicator.ack));
        }
        set ach(a) {
            this.h(BollingerBandWidthIndicator.ack, a);
        }
        ab1() {
            return this.ach;
        }
        get acg() {
            return this.c(BollingerBandWidthIndicator.acj);
        }
        set acg(a) {
            this.h(BollingerBandWidthIndicator.acj, a);
        }
        abz() {
            return this.acg;
        }
    }
    BollingerBandWidthIndicator.$t = /*@__PURE__*/ markType(BollingerBandWidthIndicator, 'BollingerBandWidthIndicator', StrategyBasedIndicator.$);
    BollingerBandWidthIndicator.ack = /*@__PURE__*/ StrategyBasedIndicator.acd(20, BollingerBandWidthIndicator.$);
    BollingerBandWidthIndicator.acj = /*@__PURE__*/ StrategyBasedIndicator.acc(2, BollingerBandWidthIndicator.$);
    return BollingerBandWidthIndicator;
})();
