// igniteui-react-charts/esm2015/lib/CalloutRenderStyleUpdatingEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let CalloutRenderStyleUpdatingEventArgs = /*@__PURE__*/ (() => {
    class CalloutRenderStyleUpdatingEventArgs extends Base {
        constructor() {
            super(...arguments);
            this._xValue = null;
            this._yValue = null;
            this._labelPositionX = 0;
            this._labelPositionY = 0;
            this._targetPositionX = 0;
            this._targetPositionY = 0;
            this._actualPosition = 0;
            this._item = null;
            this._series = null;
            this._background = null;
            this._textColor = null;
            this._outline = null;
            this._leaderBrush = null;
            this._strokeThickness = 0;
        }
        get xValue() {
            return this._xValue;
        }
        set xValue(a) {
            this._xValue = a;
        }
        get yValue() {
            return this._yValue;
        }
        set yValue(a) {
            this._yValue = a;
        }
        get labelPositionX() {
            return this._labelPositionX;
        }
        set labelPositionX(a) {
            this._labelPositionX = a;
        }
        get labelPositionY() {
            return this._labelPositionY;
        }
        set labelPositionY(a) {
            this._labelPositionY = a;
        }
        get targetPositionX() {
            return this._targetPositionX;
        }
        set targetPositionX(a) {
            this._targetPositionX = a;
        }
        get targetPositionY() {
            return this._targetPositionY;
        }
        set targetPositionY(a) {
            this._targetPositionY = a;
        }
        get actualPosition() {
            return this._actualPosition;
        }
        set actualPosition(a) {
            this._actualPosition = a;
        }
        get item() {
            return this._item;
        }
        set item(a) {
            this._item = a;
        }
        get series() {
            return this._series;
        }
        set series(a) {
            this._series = a;
        }
        get background() {
            return this._background;
        }
        set background(a) {
            this._background = a;
        }
        get textColor() {
            return this._textColor;
        }
        set textColor(a) {
            this._textColor = a;
        }
        get outline() {
            return this._outline;
        }
        set outline(a) {
            this._outline = a;
        }
        get leaderBrush() {
            return this._leaderBrush;
        }
        set leaderBrush(a) {
            this._leaderBrush = a;
        }
        get strokeThickness() {
            return this._strokeThickness;
        }
        set strokeThickness(a) {
            this._strokeThickness = a;
        }
    }
    CalloutRenderStyleUpdatingEventArgs.$t = /*@__PURE__*/ markType(CalloutRenderStyleUpdatingEventArgs, 'CalloutRenderStyleUpdatingEventArgs');
    return CalloutRenderStyleUpdatingEventArgs;
})();
