// igniteui-react-charts/esm2015/lib/CategoryTransitionInMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available transition effects that can applied to initial rendering of a series
 */
export var CategoryTransitionInMode = /*@__PURE__*/ (function (CategoryTransitionInMode) {
    /**
     * Series transitions in an automatically chosen based on type of series and its orientation
     */
    CategoryTransitionInMode[CategoryTransitionInMode["Auto"] = 0] = "Auto";
    /**
     * Series transitions in from the reference value of the value axis.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["FromZero"] = 1] = "FromZero";
    /**
     * Series sweeps in from the left
     */
    CategoryTransitionInMode[CategoryTransitionInMode["SweepFromLeft"] = 2] = "SweepFromLeft";
    /**
     * Series sweeps in from the right
     */
    CategoryTransitionInMode[CategoryTransitionInMode["SweepFromRight"] = 3] = "SweepFromRight";
    /**
     * Series sweeps in from the top.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["SweepFromTop"] = 4] = "SweepFromTop";
    /**
     * Series sweeps in from the bottom.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["SweepFromBottom"] = 5] = "SweepFromBottom";
    /**
     * Series sweeps in from the center.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["SweepFromCenter"] = 6] = "SweepFromCenter";
    /**
     * Series accordions in from the left.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["AccordionFromLeft"] = 7] = "AccordionFromLeft";
    /**
     * Series accordions in from the right.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["AccordionFromRight"] = 8] = "AccordionFromRight";
    /**
     * Series accordions in from the top.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["AccordionFromTop"] = 9] = "AccordionFromTop";
    /**
     * Series accordions in from the bottom.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["AccordionFromBottom"] = 10] = "AccordionFromBottom";
    /**
     * Series expands from the value midpoints.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["Expand"] = 11] = "Expand";
    /**
     * Series sweeps in from the category axis minimum.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["SweepFromCategoryAxisMinimum"] = 12] = "SweepFromCategoryAxisMinimum";
    /**
     * Series sweeps in from the category axis maximum.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["SweepFromCategoryAxisMaximum"] = 13] = "SweepFromCategoryAxisMaximum";
    /**
     * Series sweeps in from the value axis minimum.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["SweepFromValueAxisMinimum"] = 14] = "SweepFromValueAxisMinimum";
    /**
     * Series sweeps in from the value axis maximum.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["SweepFromValueAxisMaximum"] = 15] = "SweepFromValueAxisMaximum";
    /**
     * Series accordions in from the category axis minimum.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["AccordionFromCategoryAxisMinimum"] = 16] = "AccordionFromCategoryAxisMinimum";
    /**
     * Series accordions in from the category axis maximum.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["AccordionFromCategoryAxisMaximum"] = 17] = "AccordionFromCategoryAxisMaximum";
    /**
     * Series accordions in from the value axis minimum.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["AccordionFromValueAxisMinimum"] = 18] = "AccordionFromValueAxisMinimum";
    /**
     * Series accordions in from the value axis maximum.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["AccordionFromValueAxisMaximum"] = 19] = "AccordionFromValueAxisMaximum";
    /**
     * If the series is a child layer, animate from the values that the parent currently has.
     */
    CategoryTransitionInMode[CategoryTransitionInMode["FromParent"] = 20] = "FromParent";
    return CategoryTransitionInMode;
})({});
/**
 * @hidden
 */
export let CategoryTransitionInMode_$type = /*@__PURE__*/ markEnum('CategoryTransitionInMode', 'Auto,0|FromZero,1|SweepFromLeft,2|SweepFromRight,3|SweepFromTop,4|SweepFromBottom,5|SweepFromCenter,6|AccordionFromLeft,7|AccordionFromRight,8|AccordionFromTop,9|AccordionFromBottom,10|Expand,11|SweepFromCategoryAxisMinimum,12|SweepFromCategoryAxisMaximum,13|SweepFromValueAxisMinimum,14|SweepFromValueAxisMaximum,15|AccordionFromCategoryAxisMinimum,16|AccordionFromCategoryAxisMaximum,17|AccordionFromValueAxisMinimum,18|AccordionFromValueAxisMaximum,19|FromParent,20');
