// igniteui-react-charts/esm2015/lib/CategoryBucketCalculator_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, typeCast, markType, Array_$type, Number_$type, Point_$type, enumGetBox, runOn, delegateCombine, fromEnum, markDep } from "igniteui-react-core";
import { IBucketizer_$type } from "./IBucketizer";
import { ArgumentNullException } from "igniteui-react-core";
import { Rect } from "igniteui-react-core";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { ISortingAxis_$type } from "./ISortingAxis";
import { ScalerParams } from "./ScalerParams";
import { Series } from "./Series";
import { truncate, intDivide, isNaN_, isInfinity } from "igniteui-react-core";
import { FramePreparer } from "./FramePreparer";
import { IIsCategoryBased_$type } from "./IIsCategoryBased";
import { List$1 } from "igniteui-react-core";
import { ISupportsMarkers_$type } from "./ISupportsMarkers";
import { IProvidesViewport_$type } from "./IProvidesViewport";
import { ISupportsErrorBars_$type } from "./ISupportsErrorBars";
import { CategoryFrame } from "./CategoryFrame";
import { PreparationParams } from "./PreparationParams";
import { IDateTimeAxis_$type } from "./IDateTimeAxis";
import { NumericAxisBase } from "./NumericAxisBase";
import { IHasCategoryTrendline_$type } from "./IHasCategoryTrendline";
import { IHasSingleValueCategory_$type } from "./IHasSingleValueCategory";
import { DefaultCategoryTrendlineHost } from "./DefaultCategoryTrendlineHost";
import { DefaultSingleValueProvider } from "./DefaultSingleValueProvider";
import { SingleValuesHolder } from "./SingleValuesHolder";
import { MarkerSeries } from "./MarkerSeries";
import { IHasCategoryModePreference_$type } from "./IHasCategoryModePreference";
import { ICategorySeriesInteractionManager_$type } from "./ICategorySeriesInteractionManager";
import { DVContainer } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { CategoryCollisionMode_$type } from "./CategoryCollisionMode";
import { CategoryTransitionInMode_$type } from "./CategoryTransitionInMode";
import { HighlightingInfo } from "igniteui-react-core";
import { CategoryTransitionSourceFramePreparer } from "./CategoryTransitionSourceFramePreparer";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { Marker } from "./Marker";
import { MarkerSeriesView } from "./MarkerSeriesView";
import { CategorySeriesRenderManager } from "./CategorySeriesRenderManager";
import { Pool$1 } from "igniteui-react-core";
import { SeriesLayerPropertyOverlay } from "./SeriesLayerPropertyOverlay";
import { DoubleCollection } from "igniteui-react-core";
import { Clipper } from "igniteui-react-core";
import { ChartSelection } from "./ChartSelection";
import { PropertyMetadata } from "igniteui-react-core";
import { round10N } from "igniteui-react-core";
/**
 * @hidden
 */
export let CategoryBucketCalculator = /*@__PURE__*/ (() => {
    class CategoryBucketCalculator extends Base {
        constructor(a) {
            super();
            this.c = null;
            this.e = 0;
            this.f = 0;
            this.d = 0;
            if (a == null) {
                throw new ArgumentNullException(0, "view");
            }
            this.c = a;
            this.e = -1;
            this.d = 0;
        }
        getBucket(a) {
            return null;
        }
        getBucketWithoutUnknowns(a) {
            return this.getBucket(a);
        }
        getErrorBucket(a, b) {
            return NaN;
        }
        g(a) {
            let b = this.c.by;
            let c = this.c.bx;
            let d = this.c.cy;
            let e = typeCast(CategoryAxisBase.$, d.fetchXAxis());
            let f = d.dn;
            if (b.isEmpty || c.isEmpty || e == null || f == null || f.count == 0) {
                this.d = 0;
                return;
            }
            let g = d.getEffectiveViewport1(this.c);
            let h = d.wj(this.c);
            let i = typeCast(ISortingAxis_$type, e);
            if (i == null || i.i$b == null) {
                let j = new ScalerParams(0, b, c, e.ch, g);
                let k = Math.floor(e.eo(h.left, j));
                let l = Math.ceil(e.eo(h.right, j));
                if (e.ch) {
                    l = Math.ceil(e.eo(h.left, j));
                    k = Math.floor(e.eo(h.right, j));
                }
                l = Math.min(l, e.lk - 1);
                let m = Math.floor((l - k + 1) * a / h.width);
                this.d = truncate(Math.max(1, m));
                this.e = truncate(Math.max(0, Math.floor(k / this.d) - 1));
                this.f = truncate(Math.ceil(l / this.d));
            }
            else {
                this.e = i.i$e(b, c, g, h);
                this.f = i.i$g(b, c, g, h);
                this.d = 1;
            }
        }
        getBucketInfo(a, b, c, d) {
            a = this.e;
            b = this.f;
            c = this.d;
            d = this.c.cy.id;
            return {
                p0: a,
                p1: b,
                p2: c,
                p3: d
            };
        }
        cacheValues() {
        }
        unCacheValues() {
        }
    }
    CategoryBucketCalculator.$t = /*@__PURE__*/ markType(CategoryBucketCalculator, 'CategoryBucketCalculator', Base.$, [IBucketizer_$type]);
    return CategoryBucketCalculator;
})();
/**
 * @hidden
 */
export let CategoryFramePreparerBase = /*@__PURE__*/ (() => {
    class CategoryFramePreparerBase extends FramePreparer {
        constructor(a, ..._rest) {
            a = (a == void 0) ? 0 : a;
            switch (a) {
                case 0:
                    {
                        let c = _rest[0];
                        let d = [c, typeCast(ISupportsMarkers_$type, c), typeCast(IProvidesViewport_$type, c), typeCast(ISupportsErrorBars_$type, c), typeCast(IBucketizer_$type, c)];
                        {
                            let f = d[0];
                            let g = d[1];
                            let h = d[2];
                            let i = d[3];
                            let j = d[4];
                            super(1, g, h, i);
                            this.g = null;
                            this.h = null;
                            this.m = false;
                            this.o = new List$1(Array_$type, 0);
                            this.p = new List$1(Number_$type, 0);
                            this.n = new List$1(Number_$type, 0);
                            this.h = f;
                            this.g = j;
                        }
                    }
                    break;
                case 1:
                    {
                        let c = _rest[0];
                        let d = _rest[1];
                        let e = _rest[2];
                        let f = _rest[3];
                        let g = _rest[4];
                        super(1, d, e, f);
                        this.g = null;
                        this.h = null;
                        this.m = false;
                        this.o = new List$1(Array_$type, 0);
                        this.p = new List$1(Number_$type, 0);
                        this.n = new List$1(Number_$type, 0);
                        this.h = c;
                        this.g = g;
                    }
                    break;
            }
        }
        i(a) {
            let b = new PreparationParams();
            b.f = this.h.useHighMarkerFidelity;
            b.b = this.h.scaler;
            b.c = this.h.yScaler;
            b.d = typeCast(ISortingAxis_$type, b.b);
            b.a = typeCast(CategoryFrame.$, a);
            if (b.a == null || b.b == null || b.c == null) {
                return null;
            }
            let c;
            let d;
            let e;
            let f;
            let g = this.g.getBucketInfo(c, d, e, f);
            c = g.p0;
            d = g.p1;
            e = g.p2;
            f = g.p3;
            b.k = c;
            b.l = d;
            b.j = e;
            b.i = f;
            if (b.l < b.k) {
                return null;
            }
            let h;
            let i;
            let j;
            let k = this.a.getViewInfo(i, h);
            i = k.p0;
            h = k.p1;
            j = this.a.getEffectiveViewport();
            let l = this.a.getMainContentViewport();
            b.p = h;
            b.o = i;
            b.n = j;
            b.m = l;
            if (Rect.l_op_Equality(b.p, Rect.empty) || Rect.l_op_Equality(b.o, Rect.empty)) {
                return null;
            }
            if (this.h != null && (typeCast(Series.$, this.h) !== null && this.h.fb) && this.g != null) {
                b.e = true;
            }
            return b;
        }
        r(a, b, c, d) {
            return a.getSeriesOffset(b, c, d, this.h.currentMode2Index, this.h.currentCategoryMode);
        }
        s(a) {
            return this.r(a.b, a.p, a.o, a.n);
        }
        d(a, b) {
            let c = this.i(a);
            if (c == null || this.g == null) {
                return;
            }
            c.a.f.clear();
            c.a.k.clear();
            c.a.n.clear();
            c.a.q.clear();
            c.a.i.clear();
            c.a.h.clear();
            c.a.g.clear();
            if (c.a.m != null) {
                c.a.m.clear();
            }
            let d = this.c.shouldDisplayMarkers;
            let e = false;
            let f = this.s(c);
            let g = this.j(c);
            if (c.d != null && c.d.i$b != null && c.d.i$b.count != g.e) {
                return;
            }
            if (c.d != null && c.d.i$a) {
                c.d.i$h();
            }
            this.g.cacheValues();
            if (this.m && c.d != null && c.d.i$b.count > 0) {
                let h = c.d.i$b._inner[c.k];
                this.z(g, h, true, c.e);
                while ((isNaN_(g.c) || isNaN_(g.d)) && c.k > 0) {
                    c.k = c.k - 1;
                    h = c.d.i$b._inner[c.k];
                    this.z(g, h, true, c.e);
                }
                let i = c.d.i$b._inner[c.l];
                this.z(g, i, true, c.e);
                while ((isNaN_(g.c) || isNaN_(g.d)) && c.l < g.e - 1) {
                    c.l = c.l + 1;
                    i = c.d.i$b._inner[c.l];
                    this.z(g, i, true, c.e);
                }
            }
            if (this.m && c.d == null) {
                let j = c.k;
                let k = this.g.getBucket(j);
                while ((isNaN_(k[1]) || isNaN_(k[2])) && c.k > 0) {
                    c.k = c.k - 1;
                    j = c.k;
                    k = this.g.getBucket(j);
                }
                let l = c.l;
                k = this.g.getBucket(l);
                while ((isNaN_(k[1]) || isNaN_(k[2])) && c.l < (intDivide(g.e, c.j))) {
                    c.l = c.l + 1;
                    l = c.l;
                    k = this.g.getBucket(l);
                }
            }
            this.w(c, g, f);
            let m = this.t(c, g, f, d, e);
            this.c.updateMarkerCount(m);
            this.v(typeCast(CategoryFrame.$, a), b);
            this.g.unCacheValues();
            return;
        }
        v(a, b) {
        }
        get q() {
            return this.o;
        }
        t(a, b, c, d, e) {
            let f = 0;
            let g = false;
            let h = null;
            let i = false;
            let j = a.d != null;
            let k = a.p;
            let l = a.o;
            let m = typeCast(NumericAxisBase.$, a.c) !== null && a.c.lp;
            let n = a.f;
            let o = this.h.mayContainUnknowns;
            let p = this.o;
            let q = this.n;
            let r = this.p;
            p.clear();
            q.clear();
            r.clear();
            let s = this.g;
            let t = this.h.provideCollisionDetector();
            let u = 0;
            let v = a.n;
            let w = new ScalerParams(0, k, l, a.b.j$b, v);
            let x = ((() => {
                let $ret = new ScalerParams(0, k, l, a.c.j$b, v);
                $ret.b = typeCast(Series.$, this.h) !== null ? this.h.iy() : NaN;
                return $ret;
            })());
            if (j) {
                u = a.b.j$d(2, w) - a.b.j$d(1, w);
                i = this.k();
            }
            let y = a.k;
            let z = a.l;
            let aa = a.a.f;
            let ab = typeCast(IDateTimeAxis_$type, a.b);
            if (ab != null) {
                ab.StartTimeCursor();
            }
            for (let ac = y; ac <= z; ++ac) {
                let ad;
                if (!j) {
                    if (!o) {
                        ad = s.getBucketWithoutUnknowns(ac);
                    }
                    else {
                        ad = s.getBucket(ac);
                    }
                }
                else {
                    ad = ((() => { let ae = this.f(a, ac, b, u, g, h, c); ac = ae.p1; g = ae.p4; h = ae.p5; return ae.ret; })());
                }
                let ae = !m || (m && ad[1] > 0);
                let af = ad[0];
                if (!j || !isNaN_(ad[0])) {
                    aa.add(ad);
                    if (j) {
                        if (a.a.m == null) {
                            a.a.m = new List$1(Number_$type, 0);
                        }
                        a.a.m.add(ac * a.j);
                    }
                    if (g && i) {
                        if (h != null) {
                            aa.add(h);
                            if (j) {
                                if (a.a.m == null) {
                                    a.a.m = new List$1(Number_$type, 0);
                                }
                                a.a.m.add(ac * a.j);
                            }
                        }
                    }
                    if (d && ae) {
                        let ag = ac * a.j;
                        let ah = ag;
                        if (j && a.d.i$b != null && ag >= 0 && ag < a.d.i$b.count) {
                            ag = a.d.i$b._inner[ag];
                        }
                        let ai = new Array(ad.length);
                        ai[0] = ad[0];
                        ai[1] = ad[1];
                        ai[2] = ad[2];
                        if (n && ag < b.e) {
                            ai = new Array(ad.length);
                            ai[0] = af;
                            this.z(b, ag, true, a.e);
                            ai[1] = b.c;
                            ai[2] = b.d;
                        }
                        p.add(ai);
                        q.add(a.a.f.count - 1);
                        r.add(ag);
                    }
                }
            }
            if (ab != null) {
                ab.ResetTimeCursor();
            }
            this.y(a, c, j, w, x, e, d);
            if (d) {
                let aj = p.count;
                let ak;
                for (let al = 0; al < aj; al++) {
                    ak = p._inner[al];
                    if (this.l(a.a, ak, t, Math.min(r._inner[al], b.e - 1), f, q._inner[al])) {
                        ++f;
                    }
                }
            }
            return f;
        }
        k() {
            if (typeCast(Series.$, this.h) !== null) {
                return this.h.ev;
            }
            return false;
        }
        f(a, b, c, d, e, f, g) {
            let h;
            f = null;
            e = false;
            let i = typeCast(CategorySeries.$, this.h);
            let j = a.o;
            let k = a.p;
            let l = a.n;
            let m = new ScalerParams(0, k, j, a.b.j$b, l);
            let n = new ScalerParams(0, k, j, a.c.j$b, l);
            let o = a.e;
            if (a.d == null || a.d.i$b == null || a.d.i$b.count == 0) {
                return {
                    ret: [NaN, NaN, NaN],
                    p1: b,
                    p4: e,
                    p5: f
                };
            }
            if (i != null && i.dn != null && i.dn.count < a.d.i$b.count) {
                return {
                    ret: [NaN, NaN, NaN],
                    p1: b,
                    p4: e,
                    p5: f
                };
            }
            let p = a.d.i$b._inner[b];
            let q = a.d.i$d(p);
            let r = q;
            this.z(c, p, false, o);
            while (b < a.l) {
                p = a.d.i$b._inner[b + 1];
                r = a.d.i$d(p);
                this.z(c, p, true, o);
                if (r - q >= d || isNaN_(c.c) || isNaN_(c.d)) {
                    if (e) {
                        let s = a.d.i$b._inner[b];
                        this.z(c, s, true, o);
                        f = [a.b.j$c(q, m), c.c, c.d];
                    }
                    break;
                }
                if (!e && this.k()) {
                    let t = a.d.i$b._inner[b];
                    this.z(c, t, true, o);
                    let u = [a.b.j$c(q, m), c.c, c.d];
                    if (!isNaN_(u[0])) {
                        if (!isNaN_(u[1]) && !isNaN_(u[2])) {
                            if (a.d != null) {
                                u[0] = u[0] + g;
                            }
                            a.a.f.add(u);
                            if (a.d != null) {
                                if (a.a.m == null) {
                                    a.a.m = new List$1(Number_$type, 0);
                                }
                                a.a.m.add(b);
                            }
                            e = true;
                        }
                        else {
                            break;
                        }
                    }
                }
                b++;
                this.u(c, p, o);
            }
            let v = NaN;
            if (!isNaN_(q)) {
                v = a.b.j$c(q, m);
            }
            h = this.e(v, c);
            return {
                ret: h,
                p1: b,
                p4: e,
                p5: f
            };
        }
        y(a, b, c, d, e, f, g) {
            let h = a.j;
            let i = a.a.f.count;
            let j = a.a.f;
            let k = this.o;
            let l = k.count;
            if (f) {
                let m = a.c.j$c(a.c.me, e);
                a.c.j$f(a.a.k, 0, a.a.k.count, e);
                let n = a.a.k;
                let o = n.count;
                for (let p = 0; p < n.count; p++) {
                    n._inner[p] = Math.abs(n._inner[p] - m);
                }
            }
        }
        w(a, b, c) {
        }
    }
    CategoryFramePreparerBase.$t = /*@__PURE__*/ markType(CategoryFramePreparerBase, 'CategoryFramePreparerBase', FramePreparer.$);
    return CategoryFramePreparerBase;
})();
/**
 * @hidden
 */
export let CategoryFramePreparer = /*@__PURE__*/ (() => {
    class CategoryFramePreparer extends CategoryFramePreparerBase {
        constructor(a, ..._rest) {
            a = (a == void 0) ? 0 : a;
            switch (a) {
                case 0:
                    {
                        let c = _rest[0];
                        let d = [c, typeCast(ISupportsMarkers_$type, c), typeCast(IProvidesViewport_$type, c), typeCast(ISupportsErrorBars_$type, c), typeCast(IBucketizer_$type, c)];
                        {
                            let f = d[0];
                            let g = d[1];
                            let h = d[2];
                            let i = d[3];
                            let j = d[4];
                            super(1, f, g, h, i, j);
                            this.aa = new DefaultCategoryTrendlineHost();
                            this.ab = new DefaultSingleValueProvider();
                            this.ac = null;
                            this.ad = null;
                            this.ac = this.aa;
                            if (typeCast(IHasCategoryTrendline_$type, f) !== null) {
                                this.ac = typeCast(IHasCategoryTrendline_$type, f);
                            }
                            this.ad = this.ab;
                            if (typeCast(IHasSingleValueCategory_$type, f) !== null) {
                                this.ad = typeCast(IHasSingleValueCategory_$type, f);
                            }
                        }
                    }
                    break;
                case 1:
                    {
                        let c = _rest[0];
                        let d = _rest[1];
                        let e = _rest[2];
                        let f = _rest[3];
                        let g = _rest[4];
                        super(1, c, d, e, f, g);
                        this.aa = new DefaultCategoryTrendlineHost();
                        this.ab = new DefaultSingleValueProvider();
                        this.ac = null;
                        this.ad = null;
                        this.ac = this.aa;
                        if (typeCast(IHasCategoryTrendline_$type, c) !== null) {
                            this.ac = typeCast(IHasCategoryTrendline_$type, c);
                        }
                        this.ad = this.ab;
                        if (typeCast(IHasSingleValueCategory_$type, c) !== null) {
                            this.ad = typeCast(IHasSingleValueCategory_$type, c);
                        }
                    }
                    break;
            }
        }
        l(a, b, c, d, e, f) {
            let g = b[0];
            let h = b[1];
            let i = new Rect(0, g - 5, h - 5, 11, 11);
            if (!isNaN_(g) && !isNaN_(h) && !isInfinity(g) && !isInfinity(h) && c.tryAdd(i)) {
                a.n.add({ $type: Point_$type, x: g, y: h });
                this.c.updateMarkerTemplate(e, d, f);
                return true;
            }
            return false;
        }
        j(a) {
            let b = new SingleValuesHolder();
            let c = this.ad.valueColumn;
            b.f = c;
            return b;
        }
        x(a, b, c, d, e, f) {
            if (d) {
                b[0] = b[0] + c;
            }
            else {
                b[0] = a.b.j$c(b[0], e) + c;
            }
            b[1] = a.c.j$c(b[1], f);
            if (a.j > 1 || d) {
                b[2] = a.c.j$c(b[2], f);
            }
            else {
                b[2] = b[1];
            }
        }
        y(a, b, c, d, e, f, g) {
            super.y(a, b, c, d, e, f, g);
            let h = a.j;
            let i = a.a.f.count;
            let j = a.a.f;
            let k = this.q;
            let l = k.count;
            let m = new List$1(Number_$type, 0);
            if (!c) {
                m.add(0);
            }
            let n = new List$1(Number_$type, 0);
            n.add(1);
            if (c || h > 1) {
                n.add(2);
            }
            a.b.j$e(j, m, 0, i, d);
            a.c.j$e(j, n, 0, i, e);
            for (let o = 0; o < i; o++) {
                j._inner[o][0] = j._inner[o][0] + b;
            }
            if (h == 1 && !c) {
                for (let p = 0; p < i; p++) {
                    j._inner[p][2] = j._inner[p][1];
                }
            }
            if (g) {
                a.b.j$e(k, m, 0, l, d);
                a.c.j$e(k, n, 0, l, e);
                for (let q = 0; q < l; q++) {
                    k._inner[q][0] = k._inner[q][0] + b;
                }
                if (h == 1 && !c) {
                    for (let r = 0; r < l; r++) {
                        k._inner[r][2] = k._inner[r][1];
                    }
                }
            }
        }
        w(a, b, c) {
            if (this.ac.trendLineType == 0 || this.ac.trendlinePreparer == null || this.ac.trendLinePeriod < 1) {
                return;
            }
            this.ac.trendlinePreparer.prepareLine(this.ac, a, b, c);
        }
        z(a, b, c, d) {
            let e = a;
            let f = e.f;
            let g = f.item(b);
            let h = g;
            if (d) {
                let i = this.g.getBucket(b);
                g = i[1];
                h = i[1];
            }
            if (c) {
                e.c = g;
                e.d = h;
            }
            else {
                e.a = g;
                e.b = h;
            }
        }
        u(a, b, c) {
            let d = a;
            let e = d.f;
            if (b < e.count) {
                let f;
                if (c) {
                    let g = this.g.getBucket(b);
                    f = g[1];
                }
                else {
                    f = e.item(b);
                }
                d.a = Math.min(d.a, f);
                d.b = Math.max(d.b, f);
            }
        }
        e(a, b) {
            let c = b;
            return [a, c.a, c.b];
        }
    }
    CategoryFramePreparer.$t = /*@__PURE__*/ markType(CategoryFramePreparer, 'CategoryFramePreparer', CategoryFramePreparerBase.$);
    return CategoryFramePreparer;
})();
/**
 * @hidden
 */
export let CategorySeries = /*@__PURE__*/ (() => {
    class CategorySeries extends MarkerSeries {
        cs() {
            let a = new CategorySeriesView(this);
            return a;
        }
        get y8() {
            if (this.y7 == null) {
                this.y7 = DVContainer.instance.createInstance(ICategorySeriesInteractionManager_$type, () => this.dr.n);
            }
            return this.y7;
        }
        eb(a, b, c, d, e) {
            if (!this.hasMarkers) {
                return false;
            }
            let f = this.getEffectiveViewport1(this.cv);
            let g = this.ys();
            let h = new ScalerParams(1, this.db.v4, this.ws, g.ch);
            h.c = f;
            let i = this.y9();
            let j = new ScalerParams(1, this.db.v4, this.ws, i.ch);
            j.c = f;
            let k = this.db.v4;
            let l = this.ws;
            let m = this.currentMode2Index;
            let n = 0;
            if (g.ce && g.cc) {
                n = g.getSeriesOffset(k, l, f, m, g.categoryMode);
            }
            a.clear();
            this.w6.cj((o) => {
                if (o._visibility == 0) {
                    let p = new SeriesPointOfInterest();
                    let q = o.n;
                    let r = o.o;
                    if (q < this.ws.left || q > this.ws.right || r < this.ws.top || r > this.ws.bottom) {
                        return;
                    }
                    if (this.isVertical) {
                        if (g.cb && !g.ce && g.categoryMode != 0) {
                            p.e = o.ak;
                        }
                        else {
                            p.e = g.eo(r, h);
                        }
                        p.d = i.eo(q, j);
                    }
                    else {
                        if (g.cb && !g.ce && g.categoryMode != 0) {
                            p.d = o.ak;
                        }
                        else {
                            p.d = g.eo(q, h);
                        }
                        p.e = i.eo(r, j);
                    }
                    p.b = o.content;
                    if (b) {
                        if (this.isVertical) {
                            if (c) {
                                p.f = this.zz(p.d, d);
                            }
                            else {
                                let s = this.jd(o);
                                if (!isNaN_(s)) {
                                    p.f = this.zz(s, d);
                                }
                                else {
                                    p.f = this.zz(p.d, d);
                                }
                            }
                        }
                        else {
                            if (c) {
                                p.f = this.zz(p.e, d);
                            }
                            else {
                                let t = this.jd(o);
                                if (!isNaN_(t)) {
                                    p.f = this.zz(t, d);
                                }
                                else {
                                    p.f = this.zz(p.e, d);
                                }
                            }
                        }
                    }
                    else {
                        let u = this.au(true);
                        if (u != null) {
                            let v = this.ks(o.ak, false);
                            if (v != null) {
                                p.f = u.i(v);
                            }
                            else {
                                p.f = null;
                            }
                        }
                        else {
                            p.f = g.gf(o.ak);
                        }
                    }
                    p.c = o.ak;
                    p.g = this.vg;
                    if (this.g9) {
                        p.g = this.vn(o.ak);
                    }
                    a.add(p);
                }
            });
            return true;
        }
        zz(a, b) {
            let c;
            if (b >= 0) {
                c = round10N(a, b);
            }
            else {
                c = a;
            }
            let d = this.au(false);
            if (d != null) {
                let e = d;
                return e.i(c);
            }
            else {
                return c.toString();
            }
        }
        q1(a) {
            super.q1(a);
            this.y3 = a;
        }
        preferredCategoryMode(a) {
            return 0;
        }
        get categoryAxis() {
            return this.ys();
        }
        ys() {
            return this.fetchXAxis();
        }
        y9() {
            return this.fetchYAxis();
        }
        get_ek() {
            return true;
        }
        get_f5() {
            return this.y9() != null && this.y9().ch;
        }
        get_ey() {
            return true;
        }
        constructor() {
            super();
            this.yz = null;
            this.y7 = null;
            this.zg = false;
            this.zf = false;
            this.assigningCategoryStyle = null;
            this.assigningCategoryMarkerStyle = null;
            this.zd = null;
            this.ze = null;
            this.yw = new CategoryFrame(3);
            this.yy = new CategoryFrame(3);
            this.yu = new CategoryFrame(3);
            this.yx = new CategoryFrame(3);
            this.y2 = null;
            this.y6 = null;
            this.y2 = new CategorySeriesRenderManager();
            this.y6 = new CategoryTransitionSourceFramePreparer();
            this.zg = false;
            this.d2 = false;
            this.zf = false;
            this.ab = CategorySeries.$;
        }
        ac() {
            super.ac();
            this.y3.cw.g(this.id);
        }
        zm() {
            return (this.assigningCategoryStyle != null && this.zg) || (this.bx != 1) || this.d2 || (this.cp != 1) || (this.co != 1);
        }
        zn() {
            return (this.assigningCategoryMarkerStyle != null && this.zf) || (this.bx != 1) || this.d2 || (this.cp != 1) || (this.co != 1);
        }
        aaj(a) {
            if (this.assigningCategoryStyle != null && this.zg) {
                this.assigningCategoryStyle(this, a);
            }
        }
        aai(a) {
            if (this.assigningCategoryMarkerStyle != null && this.zf) {
                this.assigningCategoryMarkerStyle(this, a);
            }
        }
        ym(a, b) {
            return null;
        }
        o5(a, b) {
            super.o5(a, b);
            if (a) {
                this.z8(b);
            }
        }
        br() {
            let a = super.br();
            return a;
        }
        qb() {
            super.qb();
            let a = this.fetchXAxis();
            if (a != null) {
                a.ib(false);
            }
            let b = this.fetchYAxis();
            if (b != null) {
                b.ib(false);
            }
        }
        sr(a, b) {
            super.sr(a, b);
            this.y3.cw.g(this.id);
            this.rr(false);
        }
        sq(a, b) {
            if (this.g1 && this.jq < 0.05 && (this.db == null || !this.db.fn())) {
                this.g2 = true;
                this.g5 = false;
                this.gk = false;
            }
            super.sq(a, b);
            this.y3.cw.g(this.id);
            this.rr(false);
        }
        q9(a, b, c, d) {
            super.q9(a, b, c, d);
            switch (b) {
                case "SeriesViewer":
                    let e = this.fetchYAxis();
                    let f = this.fetchXAxis();
                    if (c != null && d == null) {
                        this.pc(f);
                        this.pc(e);
                    }
                    if (c == null && d != null) {
                        this.rh(f);
                        this.rh(e);
                    }
                    this.y3.cw.g(this.id);
                    this.rr(false);
                    break;
                case "SyncLink":
                    if (this.c3 != null && this.db != null) {
                        this.y3.cw.g(this.id);
                    }
                    this.rr(false);
                    break;
                case "TransitionProgress":
                    this.yy.c(this.jq, this.yw, this.yu);
                    if (this.d5(this.cv)) {
                        return;
                    }
                    if (this.jq == 1) {
                        this.yu.w();
                        this.aak(this.yu, this.y3);
                        if (this.g1) {
                            this.g1 = false;
                            this.o6();
                            this.dc.r = truncate(this.i9());
                            this.dc.b = this.dd;
                        }
                        if (this.g4) {
                            this.g4 = false;
                            this.o6();
                            this.dc.r = truncate(this.i9());
                            this.dc.b = this.dd;
                            this.q0();
                        }
                        this.qo();
                    }
                    else {
                        this.aak(this.yy, this.y3);
                    }
                    if (this.db != null) {
                        this.db.qo(this);
                    }
                    break;
                case "ErrorBarSettings":
                    this.ql();
                    break;
                case CategorySeries.$$p[5]:
                    this.rr(false);
                    break;
                case CategorySeries.$$p[2]:
                    this.zg = this.zi();
                    this.rr(false);
                    break;
                case CategorySeries.$$p[1]:
                    this.zf = this.zh();
                    this.rr(false);
                    break;
            }
        }
        zi() {
            return this.zk;
        }
        zh() {
            return this.zj;
        }
        zt(a) {
            return this.it(a, this.fetchXAxis());
        }
        get_fs() {
            return typeCast(ISortingAxis_$type, this.ys()) !== null;
        }
        a0() {
            return this.ys();
        }
        j5(a) {
            return this.y8.b$d(this, a);
        }
        kn(a) {
            return this.y8.b$f(this, a);
        }
        j3(a) {
            let b = truncate(Math.round(this.ir(a)));
            return b;
        }
        ir(a) {
            return this.y8.b$c(this, a);
        }
        dh(a, b, c) {
            if (this.fz(this.bx) && !c) {
                let d = this.zc(a, b, this.ze);
                this.ze = d;
                return d;
            }
            else {
                let e = this.za(a, b, this.ze);
                e.d = true;
                this.ze = e;
                return e;
            }
        }
        get_fe() {
            return true;
        }
        o3() {
            super.o3();
            this.zd = null;
            this.ze = null;
        }
        dg(a, b, c) {
            if (this.ed && this.fz(this.bx) && !c) {
                let d = this.zb(a, b, this.zd);
                this.zd = d;
                return d;
            }
            else {
                let e = this.za(a, b, this.zd);
                this.zd = e;
                return e;
            }
        }
        zs() {
            return this.y3.cz.a;
        }
        zc(a, b, c) {
            if (!this.xg()) {
                return null;
            }
            let d = this.yv();
            let e = this.zs();
            return this.y8.b$b(this, e, this.zw(this.cv), this.zv(this.cv), d.f, a, b, c);
        }
        zb(a, b, c) {
            let d = this.zw(this.cv);
            let e = this.zx(this.cv);
            let f = this.zv(this.cv);
            let g = this.yv();
            return this.y8.b$a(this, g.f, g.m, d, e, f, a, b, c);
        }
        za(a, b, c) {
            let d = new HighlightingInfo();
            d.i = this;
            d.g = 0;
            d.f = this.dn != null ? this.dn.count - 1 : 0;
            d.c = true;
            if (c != null && c.g == d.g && c.f == d.f) {
                return c;
            }
            return d;
        }
        yv() {
            let a = this.yu;
            if (this.d3()) {
                if (this.dc.g) {
                    this.dc.t();
                }
                a = this.yy;
            }
            return a;
        }
        aaf(a, b, c, d, e, f, g) {
            let h = this.bx != 1;
            let i = this.d2;
            let j = this.cp != 1;
            let k = this.co != 1;
            this.y2.a3(this.y2.d, d.cn, a, b, c, d, e, f, true, g);
            this.y2.aa(this.ib);
            let l = this.y2.ax;
            let m = null;
            let n = this.dn;
            let o = null;
            let p = null;
            let q = null;
            let r = null;
            let s = null;
            if (i && this.db != null && !l.e) {
                let t = this.vo(l.z);
                let u = this.vp(l.z);
                this.y2.ae = t;
                this.y2.af = u;
                l.aa = t;
                l.ab = u;
            }
            if (h && this.db != null) {
                m = this.db.dl.getHighlightingInfo(this, n, l.z, l.y, true);
                if (this.db.dl.isCrossContextHighlight(this)) {
                    o = this.db.dl.getHighlightingInfo(this, n, l.z, l.y, true);
                }
                l.c = m;
            }
            if (j && this.db != null) {
                p = this.db.dm.getHighlightingInfo(this, n, l.z, l.y, true);
                if (p == null && this.bn != null) {
                    p = this.db.dm.getHighlightingInfo(this.bn, this.bn.dn, l.z, l.y, true);
                }
                if (this.db.dm.isCrossContextHighlight(this)) {
                    r = this.db.dm.getHighlightingInfo(this, n, l.z, l.y, true);
                }
                l.d = p;
            }
            if (k && this.db != null) {
                q = this.db.dk.getHighlightingInfo(this, n, l.z, l.y, true);
                if (this.db.dk.isCrossContextHighlight(this)) {
                    s = this.db.dk.getHighlightingInfo(this, n, l.z, l.y, true);
                }
                l.b = q;
            }
            l.h = f;
            this.aai(l);
            this.y2.y(this, m, p, q, f, h, j, k, l.v, this.db.dl.totalHighWaterMark, l.x, this.db.dm.totalHighWaterMark, l.t, this.db.dk.totalHighWaterMark);
        }
        aag(a, b, c, d, e, f, g) {
            let h = this.bx != 1;
            let i = this.cp != 1;
            let j = this.co != 1;
            let k = this.d2;
            this.y2.a3(this.y2.b, d.cn, a, b, c, d, e, f, false, g);
            this.y2.ab(this.ib);
            let l = this.y2.ay;
            let m = null;
            let n = null;
            let o = null;
            let p = this.dn;
            if (k && this.db != null && !l.e) {
                let q = this.vn(l.z);
                let r = this.vq(l.z);
                this.y2.ag = q;
                this.y2.ah = r;
                l.aa = q;
                l.ab = r;
            }
            if (h && this.db != null) {
                m = this.db.dl.getHighlightingInfo(this, p, l.z, l.y, false);
                l.c = m;
            }
            if (i && this.db != null) {
                n = this.db.dm.getHighlightingInfo(this, p, l.z, l.y, false);
                if (n == null && this.bn != null) {
                    n = this.db.dm.getHighlightingInfo(this.bn, this.bn.dn, l.z, l.y, false);
                }
                l.d = n;
            }
            if (j && this.db != null) {
                o = this.db.dk.getHighlightingInfo(this, p, l.z, l.y, false);
                if (o == null && this.bn != null) {
                    o = this.db.dk.getHighlightingInfo(this.bn, this.bn.dn, l.z, l.y, false);
                }
                l.b = o;
            }
            l.h = f;
            this.aaj(l);
            this.y2.z(this, m, n, o, f, h, i, j, b < 0);
        }
        he(a, b, c) {
            let d = super.he(a, b, c);
            d = d && this.zr(this.fetchXAxis());
            d = d && this.zr(this.fetchYAxis());
            let e = c;
            if (!c.o() || b.isEmpty || a.isEmpty || this.dn == null) {
                d = false;
            }
            if (!d) {
                e.cw.d = 0;
            }
            return d;
        }
        zr(a) {
            if (a == null || a.bi == null) {
                return false;
            }
            let b = typeCast(CategoryAxisBase.$, a);
            if (b != null) {
                if (b.itemsSource == null) {
                    return false;
                }
                if (b.lj < 1) {
                    return false;
                }
            }
            else {
                let c = typeCast(NumericAxisBase.$, a);
                if (c != null) {
                    return c.l4 != c.l3;
                }
            }
            return true;
        }
        gn() {
            return this.zl;
        }
        go() {
            return true;
        }
        y4() {
            return 12;
        }
        o6() {
            super.o6();
            this.yw.w();
            this.yu.w();
            this.yy.w();
        }
        rs(a) {
            super.rs(a);
            this.y3.cw.g(this.id);
            if (this.d5(this.cv)) {
                this.yu.z();
                this.yy.z();
                this.yw.z();
                return;
            }
            if (this.gr) {
                if (this.d3()) {
                    if (this.dc.g) {
                        this.dc.t();
                    }
                    this.aak(this.yy, this.y3);
                }
                else {
                    this.aak(this.yu, this.y3);
                }
                return;
            }
            if (this.gf(a)) {
                let b = this.yw;
                if (this.d3()) {
                    if (this.dc.g) {
                        this.dc.t();
                    }
                    this.yw = this.yy;
                    this.yy = b;
                }
                else {
                    this.yw = this.yu;
                    this.yu = b;
                }
                let c = false;
                this.aah(this.yu, this.y3);
                if (this.g3) {
                    this.dc.w();
                    this.dc.r = truncate(this.ja());
                    this.dc.b = this.de != null ? this.de : this.dd;
                    if (this.g4) {
                        let d = this.yy;
                        this.yy = this.yw;
                        this.yw = d;
                        c = true;
                    }
                    this.y6.s(this, this.yw, this.yu, this.isVertical, this.fetchXAxis(), this.fetchYAxis(), this.y5, this.y4(), this.c5, this.c4(), runOn(this, this.zu), this.y3.bx);
                    if (this.g4) {
                        let e = this.yy;
                        this.yy = this.yw;
                        this.yw = e;
                    }
                    this.g4 = false;
                }
                if (this.g6) {
                    this.dc.w();
                    this.dc.r = truncate(this.jb());
                    this.dc.b = this.df != null ? this.df : this.dd;
                    if (this.g1) {
                        let f = this.yy;
                        this.yy = this.yw;
                        this.yw = f;
                        c = true;
                    }
                    this.y6.s(this, this.yw, this.yu, this.isVertical, this.fetchXAxis(), this.fetchYAxis(), this.y5, this.y4(), this.c5, this.c4(), runOn(this, this.zu), this.y3.bx);
                    let g = this.yu;
                    this.yu = this.yw;
                    this.yw = g;
                    if (this.g1) {
                        let h = this.yy;
                        this.yy = this.yw;
                        this.yw = h;
                    }
                    this.g1 = false;
                }
                this.d9 = false;
                if (!c) {
                    this.o1();
                }
                this.r7();
                if (this.g3) {
                    this.g2 = false;
                    this.g1 = true;
                }
                if (this.g6) {
                    this.g5 = false;
                    this.g4 = true;
                }
            }
            else {
                this.aah(this.yu, this.y3);
                this.aak(this.yu, this.y3);
            }
        }
        c4() {
            return 3;
        }
        c7() {
            let a = this.c4();
            switch (a) {
                case 0: return 0;
                case 1: return 1;
                case 4: return 4;
                case 2: return 2;
                case 3: return 3;
            }
            return 3;
        }
        zu() {
            if (this.isVertical) {
                return this.ws.left;
            }
            return this.ws.bottom;
        }
        zo() {
            return false;
        }
        ht() {
            let a = super.ht();
            if (this.zo()) {
                let b = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.r = true;
                    $ret.ad = "DashArray";
                    $ret.p = true;
                    $ret.x = ((() => {
                        let $ret = new DoubleCollection();
                        $ret.add(5);
                        $ret.add(5);
                        return $ret;
                    })());
                    $ret.q = true;
                    return $ret;
                })());
                a.add(b);
            }
            if (!this.zo()) {
                let c = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.r = true;
                    $ret.ad = "Brush";
                    $ret.p = true;
                    $ret.y = "ActualBrush";
                    $ret.q = true;
                    return $ret;
                })());
                c.valueResolving = delegateCombine(c.valueResolving, runOn(this, this.r0));
                a.add(c);
            }
            if (!this.zo()) {
                let d = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.r = true;
                    $ret.ad = "Outline";
                    $ret.p = true;
                    $ret.y = "ActualOutline";
                    $ret.q = true;
                    return $ret;
                })());
                d.valueResolving = delegateCombine(d.valueResolving, runOn(this, this.r0));
                a.add(d);
            }
            let e = ((() => {
                let $ret = new SeriesLayerPropertyOverlay();
                $ret.p = true;
                $ret.ad = "IsTransitionInEnabled";
                $ret.x = true;
                $ret.q = true;
                return $ret;
            })());
            a.add(e);
            let f = ((() => {
                let $ret = new SeriesLayerPropertyOverlay();
                $ret.p = true;
                $ret.ad = "TransitionInMode";
                $ret.x = enumGetBox(CategoryTransitionInMode_$type, 20);
                $ret.q = true;
                return $ret;
            })());
            a.add(f);
            return a;
        }
        kv() {
            if (this.d3()) {
                return this.bn.yy;
            }
            else {
                return this.bn.yu;
            }
        }
        zy() {
            if (this.bn != null) {
                let a = this.bn;
                return a.zy();
            }
            let b = 0;
            let c = this.fetchXAxis();
            for (let d of fromEnum(this.db.series)) {
                if (d == this) {
                    return b;
                }
                let e = typeCast(CategorySeries.$, d);
                if (e != null) {
                    let f = typeCast(CategoryAxisBase.$, e.fetchXAxis());
                    if (f == c && e.preferredCategoryMode(f) == 2) {
                        b++;
                    }
                }
            }
            return -1;
        }
        get GetErrorBarsXAxis() {
            return this.fetchXAxis();
        }
        get GetErrorBarsYAxis() {
            return this.fetchYAxis();
        }
        z9(a, b) {
            this.y2.a1(this, this.zn(), runOn(this, this.ym), this.zv(a), this.zw(a));
            let c = this.y2.d;
            let d = c != null;
            let e = this.ys();
            if (e == null) {
                return;
            }
            let f = this.getEffectiveViewport1(a);
            let g = new ScalerParams(0, a.by, a.bx, e.ch, f);
            let h = this.dn == null ? 0 : this.dn.count;
            let i = a;
            let j = b.f;
            let k = i.cw.e;
            let l = i.cw.f;
            let m = i.cw.d;
            let n = k * m;
            for (let o = 0; o < a.cz.count; o++) {
                let p = a.cz.item(o);
                let q = p.content;
                if (d) {
                    let r = p.al;
                    this.aaf(j, r, h, e, g, a.q, p.ak);
                }
                this.y2.ac(p, q);
            }
        }
        aak(a, b) {
            this.qx();
            this.wi = a.ab;
            b.aw();
        }
        aah(a, b) {
            a.v();
        }
        zv(a) {
            return a.cw.d;
        }
        zw(a) {
            return a.cw.e;
        }
        zx(a) {
            return a.cw.f;
        }
        z8(a) {
            let b = a;
            b.cz.count = 0;
        }
        rt(a, b) {
            super.rt(a, b);
            if (!this.g0) {
                this.cv.a4(b);
                return;
            }
            let c = typeCast(CategorySeriesView.$, this.cu);
            c.cw.g(this.id);
            this.cv.a4(b);
            if (this.d5(this.cu)) {
                return;
            }
            this.aal();
            this.g0 = false;
        }
        aal() {
            let a = typeCast(CategorySeriesView.$, this.cu);
            if (!this.gs) {
                this.aah(this.yx, a);
            }
            this.gs = false;
            this.aak(this.yx, a);
        }
        y0(a) {
            let b = typeCast(CategorySeriesView.$, a);
            if (b != null && b.q) {
                if (b.p) {
                    return new CategoryFramePreparer(1, typeCast(IIsCategoryBased_$type, this), typeCast(ISupportsMarkers_$type, b), b, this, b.cw);
                }
                else {
                    return new CategoryFramePreparer(1, typeCast(IIsCategoryBased_$type, this), typeCast(ISupportsMarkers_$type, b), this.db.c9.o, this, b.cw);
                }
            }
            else {
                return this.yz;
            }
        }
        aw(a) {
            if (this.fetchXAxis() == a) {
                return this.fetchYAxis();
            }
            if (this.fetchYAxis() == a) {
                return this.fetchXAxis();
            }
            return null;
        }
        ai() {
            return [this.fetchXAxis(), this.fetchYAxis()];
        }
        static yn(a, b, c, d, e) {
            let f = null;
            if (b > -1 && !d.isEmpty && !c.isEmpty) {
                let g = a._inner[0][0] < c.left - 2000 ? c.left - 10 : NaN;
                let h = c.bottom + 10;
                let i = a._inner[b][0] > c.right + 2000 ? c.right + 10 : NaN;
                let j = c.top - 10;
                f = new Clipper(1, g, h, i, j, e);
            }
            return f;
        }
        aac(a, b, c, d, e) {
            let f = this.wr(a);
            let g = this.ys();
            let h = this.y9();
            let i = this.fetchXAxis();
            let j = g == i;
            let k = this.wj(this.cv);
            let l, m;
            let n = this.aab(f, g, j, k, l, m);
            l = n.p4;
            m = n.p5;
            if (d == 0) {
                this.aae(b, l, m, c);
            }
            else {
                let o, p;
                let q = this.aaa(f, h, j, k, o, p);
                o = q.p4;
                p = q.p5;
                this.aad(e, b, l, m, o, p, c);
            }
        }
        aae(a, b, c, d) {
            if (b == -1 || c == -1) {
                return;
            }
            if (d == 7 || d == 6) {
                a.add(((() => {
                    let $ret = new ChartSelection();
                    $ret.series = this;
                    return $ret;
                })()));
                return;
            }
            for (let e = b; e <= c; e++) {
                this.z7(a, e, d);
            }
        }
        aaa(a, b, c, d, e, f) {
            let g = new ScalerParams(0, a, this.db.wk, b.ch, this.getEffectiveViewport());
            let h = c ? (b.ch ? d.top : d.bottom) : (b.ch ? d.right : d.left);
            let i = c ? (b.ch ? d.bottom : d.top) : (b.ch ? d.left : d.right);
            e = b.eo(h, g);
            f = b.eo(i, g);
            return {
                p4: e,
                p5: f
            };
        }
        aab(a, b, c, d, e, f) {
            let g = new ScalerParams(0, a, this.db.wk, b.ch, this.getEffectiveViewport());
            let h = c ? (b.ch ? d.right : d.left) : (b.ch ? d.top : d.bottom);
            let i = c ? (b.ch ? d.left : d.right) : (b.ch ? d.bottom : d.top);
            let j = b.eo(h, g);
            let k = b.eo(i, g);
            e = Math.max(0, truncate((b.categoryMode == 0 ? Math.ceil(j) : Math.floor(j))));
            f = Math.min(truncate(Math.floor(k)), this.dn.count - 1);
            return {
                p4: e,
                p5: f
            };
        }
    }
    CategorySeries.$t = /*@__PURE__*/ markType(CategorySeries, 'CategorySeries', MarkerSeries.$, [IHasCategoryModePreference_$type, ISupportsErrorBars_$type]);
    CategorySeries.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, CategorySeries, 'raisePropertyChanged', ['CategoryCollisionMode:yt:aam', [CategoryCollisionMode_$type, /*@__PURE__*/ enumGetBox(CategoryCollisionMode_$type, 0)], 'IsCustomCategoryMarkerStyleAllowed:zj:aan', [0, false], 'IsCustomCategoryStyleAllowed:zk:aao', [0, false], 'IsTransitionInEnabled:zl:aap', [0, false], 'TransitionInMode:y5:aaq', [CategoryTransitionInMode_$type, /*@__PURE__*/ enumGetBox(CategoryTransitionInMode_$type, 0)], 'UseHighMarkerFidelity:zq:aar', [0, false]]);
    return CategorySeries;
})();
/**
 * @hidden
 */
export let CategorySeriesView = /*@__PURE__*/ (() => {
    class CategorySeriesView extends MarkerSeriesView {
        c0(a) {
            if (this.c1 != a.u) {
                return true;
            }
            return false;
        }
        c3(a) {
            this.c1 = a.u;
        }
        constructor(a) {
            super(a);
            this.cy = null;
            this.c1 = 0;
            this.cw = null;
            this.cz = null;
            this.c1 = -1;
            this.cy = a;
            this.cw = this.cx();
            this.cz = ((() => {
                let $ret = new Pool$1(Marker.$);
                $ret.create = runOn(this, this.b3);
                $ret.activate = runOn(this, this.co);
                $ret.disactivate = runOn(this, this.cq);
                $ret.destroy = runOn(this, this.cp);
                return $ret;
            })());
        }
        updateMarkerTemplate(a, b, c) {
            if (!this.b4.xi) {
                let d = this.cz.item(a);
                let e = d.content;
                e.item = this.e.dn.item(b);
                d.ak = b;
                d.al = c;
            }
        }
        cx() {
            return new CategoryBucketCalculator(this);
        }
        get shouldDisplayMarkers() {
            return this.cy.xg();
        }
        updateMarkerCount(a) {
            this.cz.count = a;
        }
        cj(a) {
            this.cz.c(a);
        }
        c2() {
        }
    }
    CategorySeriesView.$t = /*@__PURE__*/ markType(CategorySeriesView, 'CategorySeriesView', MarkerSeriesView.$, [ISupportsMarkers_$type]);
    return CategorySeriesView;
})();
