// igniteui-react-charts/esm2015/lib/ChaikinVolatilityIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ChaikinVolatilityIndicatorStrategy } from "./ChaikinVolatilityIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ChaikinVolatilityIndicator = /*@__PURE__*/ (() => {
    class ChaikinVolatilityIndicator extends StrategyBasedIndicator {
        bj() {
            return new ChaikinVolatilityIndicator();
        }
        get_abv() {
            return new ChaikinVolatilityIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return ChaikinVolatilityIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(ChaikinVolatilityIndicator.ach));
        }
        set acg(a) {
            this.h(ChaikinVolatilityIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    ChaikinVolatilityIndicator.$t = /*@__PURE__*/ markType(ChaikinVolatilityIndicator, 'ChaikinVolatilityIndicator', StrategyBasedIndicator.$);
    ChaikinVolatilityIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(10, ChaikinVolatilityIndicator.$);
    return ChaikinVolatilityIndicator;
})();
