// igniteui-react-charts/esm2015/lib/ChaikinOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { ChaikinOscillatorIndicatorStrategy } from "./ChaikinOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ChaikinOscillatorIndicator = /*@__PURE__*/ (() => {
    class ChaikinOscillatorIndicator extends StrategyBasedIndicator {
        bj() {
            return new ChaikinOscillatorIndicator();
        }
        get_abv() {
            return new ChaikinOscillatorIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return ChaikinOscillatorIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get shortPeriod() {
            return typeGetValue(this.c(ChaikinOscillatorIndicator.ach));
        }
        set shortPeriod(a) {
            this.h(ChaikinOscillatorIndicator.ach, a);
        }
        ab3() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(ChaikinOscillatorIndicator.acg));
        }
        set longPeriod(a) {
            this.h(ChaikinOscillatorIndicator.acg, a);
        }
        ab0() {
            return this.longPeriod;
        }
    }
    ChaikinOscillatorIndicator.$t = /*@__PURE__*/ markType(ChaikinOscillatorIndicator, 'ChaikinOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    ChaikinOscillatorIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acf(3, ChaikinOscillatorIndicator.$);
    ChaikinOscillatorIndicator.acg = /*@__PURE__*/ StrategyBasedIndicator.acb(10, ChaikinOscillatorIndicator.$);
    return ChaikinOscillatorIndicator;
})();
