// igniteui-react-charts/esm2015/lib/ChartSortDescriptionCollection.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { SyncableObservableCollection$2 } from "igniteui-react-core";
import { ChartSortDescription } from "./ChartSortDescription";
import { SortDescription } from "igniteui-react-core";
/**
 * @hidden
 */
export let ChartSortDescriptionCollection = /*@__PURE__*/ (() => {
    class ChartSortDescriptionCollection extends Base {
        get i() {
            return this.f.shouldDetachOnTargetChange;
        }
        set i(a) {
            this.f.shouldDetachOnTargetChange = a;
        }
        constructor() {
            super();
            this.f = new SyncableObservableCollection$2(ChartSortDescription.$, SortDescription.$, 0);
            this.d = null;
            let a = this.f;
            a.compare = (b, c) => {
                if (b.f == c.propertyName) {
                    return true;
                }
                return false;
            };
            a.createFrom = (b) => new ChartSortDescription(2, b.propertyName, b.direction);
            a.createTo = (b) => {
                let c = b.c;
                return SortDescription.create(b.f, c, null);
            };
        }
        get e() {
            return this.d;
        }
        set e(a) {
            if (a == null) {
                this.f.syncTarget = null;
                this.d = null;
                return;
            }
            this.d = a;
            this.f.syncTarget = this.d.getInner();
        }
        g(a) {
            return this.f.add(a);
        }
        n(a) {
            this.f.addOneWayTarget(a.getInner());
        }
        r(a) {
            this.f.removeOneWayTarget(a.getInner());
        }
        p(a, b) {
            this.f.add1(a, b);
        }
        o() {
            this.f.clear();
        }
        get(a) {
            return this.f.get(a);
        }
        k(a) {
            return this.f.indexOf(a);
        }
        h(a) {
            return this.f.remove1(a);
        }
        b(a) {
            return this.f.remove(a);
        }
        set(a, b) {
            return this.f.set(a, b);
        }
        size() {
            return this.f.size();
        }
        m(a) {
            this.f.addListener(a);
        }
        q(a) {
            this.f.removeListener(a);
        }
        get j() {
            return this.f.all;
        }
    }
    ChartSortDescriptionCollection.$t = /*@__PURE__*/ markType(ChartSortDescriptionCollection, 'ChartSortDescriptionCollection');
    return ChartSortDescriptionCollection;
})();
