// igniteui-react-charts/esm2015/lib/ChartSortDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ChartSortDescription = /*@__PURE__*/ (() => {
    class ChartSortDescription extends Base {
        constructor(a, ..._rest) {
            super();
            this.e = null;
            this.b = 0;
            a = (a == void 0) ? 0 : a;
            switch (a) {
                case 0: break;
                case 1:
                    {
                        let c = _rest[0];
                        this.f = c;
                    }
                    break;
                case 2:
                    {
                        let c = _rest[0];
                        let d = _rest[1];
                        this.f = c;
                        this.c = d;
                    }
                    break;
            }
        }
        get f() {
            return this.e;
        }
        set f(a) {
            this.e = a;
        }
        get c() {
            return this.b;
        }
        set c(a) {
            this.b = a;
        }
        equals(a) {
            let b = a;
            if (b == null) {
                return super.equals(a);
            }
            return b.f == this.f && b.c == this.c;
        }
        getHashCode() {
            return Base.getHashCodeStatic(this.f) * 17 + this.c;
        }
    }
    ChartSortDescription.$t = /*@__PURE__*/ markType(ChartSortDescription, 'ChartSortDescription');
    return ChartSortDescription;
})();
