// igniteui-react-charts/esm2015/lib/CommodityChannelIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { CommodityChannelIndexIndicatorStrategy } from "./CommodityChannelIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let CommodityChannelIndexIndicator = /*@__PURE__*/ (() => {
    class CommodityChannelIndexIndicator extends StrategyBasedIndicator {
        bj() {
            return new CommodityChannelIndexIndicator();
        }
        get_abv() {
            return new CommodityChannelIndexIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return CommodityChannelIndexIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(CommodityChannelIndexIndicator.ach));
        }
        set acg(a) {
            this.h(CommodityChannelIndexIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    CommodityChannelIndexIndicator.$t = /*@__PURE__*/ markType(CommodityChannelIndexIndicator, 'CommodityChannelIndexIndicator', StrategyBasedIndicator.$);
    CommodityChannelIndexIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(20, CommodityChannelIndexIndicator.$);
    return CommodityChannelIndexIndicator;
})();
