// igniteui-react-charts/esm2015/lib/ContainerResizedMessage_Spark.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Message_Spark } from "./Message_Spark";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ContainerResizedMessage_Spark = /*@__PURE__*/ (() => {
    class ContainerResizedMessage_Spark extends Message_Spark {
        constructor() {
            super(...arguments);
            this.g = 0;
            this.d = 0;
            this.f = 0;
            this.e = 0;
            this.h = null;
            this.i = null;
            this.j = null;
            this.k = null;
            this.b = null;
            this.c = null;
        }
    }
    ContainerResizedMessage_Spark.$t = /*@__PURE__*/ markType(ContainerResizedMessage_Spark, 'ContainerResizedMessage_Spark', Message_Spark.$);
    return ContainerResizedMessage_Spark;
})();
