// igniteui-react-charts/esm2015/lib/CrosshairsDisplayMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * An enum representing the available display modes of crosshairs lines in the chart.
 */
export var CrosshairsDisplayMode = /*@__PURE__*/ (function (CrosshairsDisplayMode) {
    /**
     * Display crosshair lines depending on chart type, e.g. no vertical crosshair lines for column and waterfall charts.
     */
    CrosshairsDisplayMode[CrosshairsDisplayMode["Auto"] = 0] = "Auto";
    /**
     * Display no crosshair lines in the chart.
     */
    CrosshairsDisplayMode[CrosshairsDisplayMode["None"] = 1] = "None";
    /**
     * Display horizontal line of crosshairs in the chart.
     */
    CrosshairsDisplayMode[CrosshairsDisplayMode["Horizontal"] = 2] = "Horizontal";
    /**
     * Display vertical line of crosshairs in the chart.
     */
    CrosshairsDisplayMode[CrosshairsDisplayMode["Vertical"] = 3] = "Vertical";
    /**
     * Display both horizontal and vertical lines of crosshairs in the chart.
     */
    CrosshairsDisplayMode[CrosshairsDisplayMode["Both"] = 4] = "Both";
    return CrosshairsDisplayMode;
})({});
/**
 * @hidden
 */
export let CrosshairsDisplayMode_$type = /*@__PURE__*/ markEnum('CrosshairsDisplayMode', 'Auto,0|None,1|Horizontal,2|Vertical,3|Both,4');
