// igniteui-react-charts/esm2015/lib/CustomContourValueResolverEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { EventArgs, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let CustomContourValueResolverEventArgs = /*@__PURE__*/ (() => {
    class CustomContourValueResolverEventArgs extends EventArgs {
        constructor(a) {
            super();
            this._valueColumn = null;
            this._contourValuesEnumerable = null;
            this.b = null;
            this.valueColumn = a;
        }
        get valueColumn() {
            return this._valueColumn;
        }
        set valueColumn(a) {
            this._valueColumn = a;
        }
        get contourValuesEnumerable() {
            return this._contourValuesEnumerable;
        }
        set contourValuesEnumerable(a) {
            this._contourValuesEnumerable = a;
        }
        get values() {
            if (this.valueColumn == null) {
                return [];
            }
            return this.valueColumn.asArray();
        }
        get contourValues() {
            return this.b;
        }
        set contourValues(a) {
            if (a == null) {
                this.contourValuesEnumerable = null;
            }
            else {
                this.contourValuesEnumerable = a;
            }
        }
    }
    CustomContourValueResolverEventArgs.$t = /*@__PURE__*/ markType(CustomContourValueResolverEventArgs, 'CustomContourValueResolverEventArgs', EventArgs.$);
    return CustomContourValueResolverEventArgs;
})();
