// igniteui-react-charts/esm2015/lib/CustomIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { CustomIndicatorStrategy } from "./CustomIndicatorStrategy";
import { String_$type, fromEnum, markType } from "igniteui-react-core";
import { FinancialEventArgs } from "./FinancialEventArgs";
import { List$1 } from "igniteui-react-core";
/**
 * @hidden
 */
export let CustomIndicator = /*@__PURE__*/ (() => {
    class CustomIndicator extends StrategyBasedIndicator {
        constructor() {
            super(...arguments);
            this.ach = null;
            this.acg = null;
        }
        bj() {
            return new CustomIndicator();
        }
        get_abv() {
            return new CustomIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return CustomIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        aa7(a, b) {
            let c = this.xj(a, b);
            let d = this.xk(c);
            let e = new List$1(String_$type, 0);
            if (this.basedOnColumns != null) {
                let f = new FinancialEventArgs(c.calculateFrom, c.calculateCount, c, d, this);
                this.basedOnColumns(this, f);
                if (f.basedOn != null) {
                    for (let g of fromEnum(f.basedOn)) {
                        e.add(g);
                    }
                }
            }
            return e;
        }
        get indicator() {
            return this.ach;
        }
        set indicator(a) {
            this.ach = a;
            this.abf();
        }
        get basedOnColumns() {
            return this.acg;
        }
        set basedOnColumns(a) {
            this.acg = a;
        }
        aa4(a, b) {
            super.aa4(a, b);
            if (this.indicator != null) {
                let c = this.xj(a, b);
                if (b == 0) {
                    return false;
                }
                if (!this.yf(this.aa7(a, b))) {
                    return false;
                }
                this.indicator(this, new FinancialEventArgs(a, b, c, this.xk(c), this));
                if (this.abx(c) && this.xp != null) {
                    this.xp.c1();
                }
                return true;
            }
            return false;
        }
    }
    CustomIndicator.$t = /*@__PURE__*/ markType(CustomIndicator, 'CustomIndicator', StrategyBasedIndicator.$);
    return CustomIndicator;
})();
