// igniteui-react-charts/esm2015/lib/DataLegendSeriesGroupInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Number_$type, markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { IDataLegendSeries_$type } from "igniteui-react-core";
import { DataLegendSeriesContext } from "igniteui-react-core";
import { Dictionary$2 } from "igniteui-react-core";
import { DataLegendColumn } from "./DataLegendColumn";
/**
 * @hidden
 */
export let DataLegendSeriesGroupInfo = /*@__PURE__*/ (() => {
    class DataLegendSeriesGroupInfo extends Base {
        constructor() {
            super();
            this.c = null;
            this.b = null;
            this.a = null;
            this.f = null;
            this.d = 0;
            this.e = 0;
            this.c = new List$1(IDataLegendSeries_$type, 0);
            this.b = new List$1(DataLegendSeriesContext.$, 0);
            this.a = new Dictionary$2(Number_$type, DataLegendColumn.$, 0);
            this.d = 0;
        }
    }
    DataLegendSeriesGroupInfo.$t = /*@__PURE__*/ markType(DataLegendSeriesGroupInfo, 'DataLegendSeriesGroupInfo');
    return DataLegendSeriesGroupInfo;
})();
