// igniteui-react-charts/esm2015/lib/DataLegendSummaryResult.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Number_$type, markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
/**
 * @hidden
 */
export let DataLegendSummaryResult = /*@__PURE__*/ (() => {
    class DataLegendSummaryResult extends Base {
        constructor(a, ..._rest) {
            super();
            this.c = 0;
            this.f = null;
            this.e = null;
            this.d = null;
            this.a = null;
            a = (a == void 0) ? 0 : a;
            switch (a) {
                case 0:
                    {
                        this.c = NaN;
                        this.a = new List$1(Number_$type, 0).toArray();
                        this.d = "";
                    }
                    break;
                case 1:
                    {
                        let c = _rest[0];
                        let d = _rest[1];
                        let e = _rest[2];
                        this.c = c;
                        this.f = null;
                        this.a = d;
                        this.d = e;
                    }
                    break;
            }
        }
    }
    DataLegendSummaryResult.$t = /*@__PURE__*/ markType(DataLegendSummaryResult, 'DataLegendSummaryResult');
    DataLegendSummaryResult.b = /*@__PURE__*/ new DataLegendSummaryResult(0);
    return DataLegendSummaryResult;
})();
