// igniteui-react-charts/esm2015/lib/DataToolTipLayerFrame.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Frame } from "./Frame";
import { List$1 } from "igniteui-react-core";
import { Base, Number_$type, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
import { FrameExtended } from "./FrameExtended";
import { InterpolationUtil } from "igniteui-react-core";
/**
 * @hidden
 */
export let DataToolTipLayerFrame = /*@__PURE__*/ (() => {
    class DataToolTipLayerFrame extends Frame {
        constructor() {
            super();
            this.m = null;
            this.l = null;
            this.g = null;
            this.f = null;
            this.i = null;
            this.k = null;
            this.j = null;
            this.h = null;
            this.m = new List$1(Number_$type, 0);
            this.l = new List$1(Number_$type, 0);
            this.g = new List$1(Number_$type, 0);
            this.f = new List$1(Number_$type, 0);
            this.i = new List$1(Base.$, 0);
            this.k = new List$1(Number_$type, 0);
            this.j = new List$1(Number_$type, 0);
            this.h = new List$1(Brush.$, 0);
        }
        c(a, b, c) {
            let d = b;
            let e = c;
            Frame.b(this.m, a, d.m, e.m);
            Frame.b(this.l, a, d.l, e.l);
            Frame.b(this.g, a, d.g, e.g);
            Frame.b(this.f, a, d.f, e.f);
            Frame.b(this.k, a, d.k, e.k);
            Frame.b(this.j, a, d.j, e.j);
            FrameExtended.a(this.h, a, d.h, e.h, 0);
            InterpolationUtil.d(Base.$, this.i, a, d.i, e.i, () => null, (f, g, h, i) => i);
        }
    }
    DataToolTipLayerFrame.$t = /*@__PURE__*/ markType(DataToolTipLayerFrame, 'DataToolTipLayerFrame', Frame.$);
    return DataToolTipLayerFrame;
})();
