// igniteui-react-charts/esm2015/lib/DataToolTipSeriesInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let DataToolTipSeriesInfo = /*@__PURE__*/ (() => {
    class DataToolTipSeriesInfo extends Base {
        constructor() {
            super(...arguments);
            this.f = null;
            this.b = null;
            this.e = null;
            this.a = false;
            this.d = 0;
            this.c = 0;
        }
    }
    DataToolTipSeriesInfo.$t = /*@__PURE__*/ markType(DataToolTipSeriesInfo, 'DataToolTipSeriesInfo');
    return DataToolTipSeriesInfo;
})();
