// igniteui-react-charts/esm2015/lib/DataUpdatedMessage.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ConfigurationMessage } from "./ConfigurationMessage";
import { ItemsSourceAction_$type } from "./ItemsSourceAction";
import { EnumUtil, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let DataUpdatedMessage = /*@__PURE__*/ (() => {
    class DataUpdatedMessage extends ConfigurationMessage {
        constructor() {
            super(...arguments);
            this.b = 0;
            this.e = 0;
            this.d = 0;
            this.h = null;
        }
        get c() {
            return this.b;
        }
        set c(a) {
            this.b = a;
        }
        get g() {
            return this.e;
        }
        set g(a) {
            this.e = a;
        }
        get f() {
            return this.d;
        }
        set f(a) {
            this.d = a;
        }
        get i() {
            return this.h;
        }
        set i(a) {
            this.h = a;
        }
        toString() {
            return "DataUpdatedMessage[" + EnumUtil.getName(ItemsSourceAction_$type, this.c) + "," + this.g.toString() + ", " + this.f.toString() + ", " + this.i.toString() + "]";
        }
    }
    DataUpdatedMessage.$t = /*@__PURE__*/ markType(DataUpdatedMessage, 'DataUpdatedMessage', ConfigurationMessage.$);
    return DataUpdatedMessage;
})();
