// igniteui-react-charts/esm2015/lib/DataAbbreviation.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { isNaN_, isInfinity } from "igniteui-react-core";
/**
 * @hidden
 */
export let DataAbbreviation = /*@__PURE__*/ (() => {
    class DataAbbreviation extends Base {
        constructor() {
            super();
            this.g = 0;
            this.m = null;
            this.k = null;
            this.d = 0;
            this.j = null;
            this.n = null;
            this.f = 0;
            this.e = 0;
            this.a = 0;
            this.l = null;
            this.c = 0;
            this.i = 0;
            this.h = 0;
            this.a = 9;
            this.c = 1;
            this.l = "";
            this.j = "";
            this.m = "N/A";
            this.g = NaN;
            this.i = 5;
            this.h = -5;
            this.d = NaN;
        }
        b() {
            return !isNaN_(this.d) && !isInfinity(this.d);
        }
    }
    DataAbbreviation.$t = /*@__PURE__*/ markType(DataAbbreviation, 'DataAbbreviation');
    return DataAbbreviation;
})();
