// igniteui-react-charts/esm2015/lib/DataLegendColumn.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Number_$type, String_$type, markType } from "igniteui-react-core";
import { DataLegendIndexDictionary$1 } from "./DataLegendIndexDictionary$1";
import { DataAbbreviation } from "./DataAbbreviation";
import { DataLegendSeriesValueInfo } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
/**
 * @hidden
 */
export let DataLegendColumn = /*@__PURE__*/ (() => {
    class DataLegendColumn extends Base {
        constructor() {
            super();
            this.g = false;
            this.i = 0;
            this.f = null;
            this.e = null;
            this.b = null;
            this.d = null;
            this.a = null;
            this.c = null;
            this.h = null;
            this.h = new List$1(Number_$type, 0);
            this.e = new DataLegendIndexDictionary$1(Number_$type);
            this.f = new DataLegendIndexDictionary$1(String_$type);
            this.d = new DataLegendIndexDictionary$1(String_$type);
            this.b = new DataLegendIndexDictionary$1(String_$type);
            this.d = new DataLegendIndexDictionary$1(String_$type);
            this.a = new DataLegendIndexDictionary$1(DataAbbreviation.$);
            this.c = new DataLegendIndexDictionary$1(DataLegendSeriesValueInfo.$);
            this.g = false;
        }
    }
    DataLegendColumn.$t = /*@__PURE__*/ markType(DataLegendColumn, 'DataLegendColumn');
    return DataLegendColumn;
})();
