// igniteui-react-charts/esm2015/lib/DataLegendStylingRowEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let DataLegendStylingRowEventArgs = /*@__PURE__*/ (() => {
    class DataLegendStylingRowEventArgs extends Base {
        constructor() {
            super();
            this.a = 0;
            this.b = false;
            this.c = false;
            this.d = 0;
            this.f = null;
            this.g = null;
            this.e = null;
            this.h = null;
            this.b = true;
            this.c = true;
            this.d = -1;
            this.a = 7;
        }
    }
    DataLegendStylingRowEventArgs.$t = /*@__PURE__*/ markType(DataLegendStylingRowEventArgs, 'DataLegendStylingRowEventArgs');
    return DataLegendStylingRowEventArgs;
})();
