// igniteui-react-charts/esm2015/lib/DataToolTipLayerPosition.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available positions on the crossing axis at which the category tooltip can be displayed.
 */
export var DataToolTipLayerPosition = /*@__PURE__*/ (function (DataToolTipLayerPosition) {
    /**
     * A position is selected automatically
     */
    DataToolTipLayerPosition[DataToolTipLayerPosition["Auto"] = 0] = "Auto";
    /**
     * Displays at the outside start of the axis.
     */
    DataToolTipLayerPosition[DataToolTipLayerPosition["OutsideStart"] = 1] = "OutsideStart";
    /**
     * Displays at the inside start of the axis.
     */
    DataToolTipLayerPosition[DataToolTipLayerPosition["InsideStart"] = 2] = "InsideStart";
    /**
     * Displays at the inside end of the axis.
     */
    DataToolTipLayerPosition[DataToolTipLayerPosition["InsideEnd"] = 3] = "InsideEnd";
    /**
     * Displays at the outside end of the axis.
     */
    DataToolTipLayerPosition[DataToolTipLayerPosition["OutsideEnd"] = 4] = "OutsideEnd";
    return DataToolTipLayerPosition;
})({});
/**
 * @hidden
 */
export let DataToolTipLayerPosition_$type = /*@__PURE__*/ markEnum('DataToolTipLayerPosition', 'Auto,0|OutsideStart,1|InsideStart,2|InsideEnd,3|OutsideEnd,4');
