// igniteui-react-charts/esm2015/lib/DataLegendSummaryEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { EventArgs, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let DataLegendSummaryEventArgs = /*@__PURE__*/ (() => {
    class DataLegendSummaryEventArgs extends EventArgs {
        constructor(a, b) {
            super();
            this._columnValues = null;
            this._columnMemberPath = null;
            this._summaryValue = 0;
            this._summaryUnits = null;
            this._summaryLabel = null;
            this.columnValues = a;
            this.columnMemberPath = b;
            this.summaryLabel = null;
            this.summaryUnits = null;
        }
        get columnValues() {
            return this._columnValues;
        }
        set columnValues(a) {
            this._columnValues = a;
        }
        get columnMemberPath() {
            return this._columnMemberPath;
        }
        set columnMemberPath(a) {
            this._columnMemberPath = a;
        }
        get summaryValue() {
            return this._summaryValue;
        }
        set summaryValue(a) {
            this._summaryValue = a;
        }
        get summaryUnits() {
            return this._summaryUnits;
        }
        set summaryUnits(a) {
            this._summaryUnits = a;
        }
        get summaryLabel() {
            return this._summaryLabel;
        }
        set summaryLabel(a) {
            this._summaryLabel = a;
        }
    }
    DataLegendSummaryEventArgs.$t = /*@__PURE__*/ markType(DataLegendSummaryEventArgs, 'DataLegendSummaryEventArgs', EventArgs.$);
    return DataLegendSummaryEventArgs;
})();
