// igniteui-react-charts/esm2015/lib/DefaultCategoryTrendlineHost.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { IHasCategoryTrendline_$type } from "./IHasCategoryTrendline";
import { DefaultCategoryTrendlinePreparer } from "./DefaultCategoryTrendlinePreparer";
/**
 * @hidden
 */
export let DefaultCategoryTrendlineHost = /*@__PURE__*/ (() => {
    class DefaultCategoryTrendlineHost extends Base {
        constructor() {
            super();
            this.a = null;
            this.a = new DefaultCategoryTrendlinePreparer();
        }
        get trendlinePreparer() {
            return this.a;
        }
        get trendLineType() {
            return 0;
        }
        get trendLinePeriod() {
            return 1;
        }
        get isVertical() {
            return false;
        }
    }
    DefaultCategoryTrendlineHost.$t = /*@__PURE__*/ markType(DefaultCategoryTrendlineHost, 'DefaultCategoryTrendlineHost', Base.$, [IHasCategoryTrendline_$type]);
    return DefaultCategoryTrendlineHost;
})();
