// igniteui-react-charts/esm2015/lib/DefaultPaletteInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let DefaultPaletteInfo = /*@__PURE__*/ (() => {
    class DefaultPaletteInfo extends Base {
        constructor() {
            super(...arguments);
            this.a = null;
            this.f = null;
            this.b = null;
            this.c = null;
            this.g = null;
            this.d = null;
            this.e = null;
            this.k = null;
            this.j = null;
            this.h = null;
            this.i = null;
        }
    }
    DefaultPaletteInfo.$t = /*@__PURE__*/ markType(DefaultPaletteInfo, 'DefaultPaletteInfo');
    return DefaultPaletteInfo;
})();
