// igniteui-react-charts/esm2015/lib/DefaultProvidesViewport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { IProvidesViewport_$type } from "./IProvidesViewport";
import { Rect } from "igniteui-react-core";
/**
 * @hidden
 */
export let DefaultProvidesViewport = /*@__PURE__*/ (() => {
    class DefaultProvidesViewport extends Base {
        getViewInfo(a, b) {
            a = Rect.empty;
            b = Rect.empty;
            return {
                p0: a,
                p1: b
            };
        }
        getEffectiveViewport() {
            return Rect.empty;
        }
        getMainContentViewport() {
            return Rect.empty;
        }
    }
    DefaultProvidesViewport.$t = /*@__PURE__*/ markType(DefaultProvidesViewport, 'DefaultProvidesViewport', Base.$, [IProvidesViewport_$type]);
    return DefaultProvidesViewport;
})();
