// igniteui-react-charts/esm2015/lib/DomainChartSeriesPointerEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Point_$type, markType } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
export let DomainChartSeriesPointerEventArgs = /*@__PURE__*/ (() => {
    class DomainChartSeriesPointerEventArgs extends Base {
        constructor(a, b, c) {
            super();
            this.e = { $type: Point_$type, x: NaN, y: NaN };
            this._series = null;
            this._item = null;
            this._plotAreaPosition = null;
            this.b = false;
            this.series = a;
            this.item = b;
            this.plotAreaPosition = c;
        }
        get series() {
            return this._series;
        }
        set series(a) {
            this._series = a;
        }
        get item() {
            return this._item;
        }
        set item(a) {
            this._item = a;
        }
        get plotAreaPosition() {
            return this._plotAreaPosition;
        }
        set plotAreaPosition(a) {
            this._plotAreaPosition = a;
        }
        get chartPosition() {
            if (isNaN_(this.e.x) && isNaN_(this.e.y)) {
                this.e.x = this.plotAreaPosition.x - this.series.c9.wk.left;
                this.e.y = this.plotAreaPosition.y - this.series.c9.wk.top;
            }
            return this.e;
        }
        get cancelSelection() {
            return this.b;
        }
        set cancelSelection(a) {
            this.b = a;
        }
    }
    DomainChartSeriesPointerEventArgs.$t = /*@__PURE__*/ markType(DomainChartSeriesPointerEventArgs, 'DomainChartSeriesPointerEventArgs');
    return DomainChartSeriesPointerEventArgs;
})();
