// igniteui-react-charts/esm2015/lib/FastStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { FastStochasticOscillatorIndicatorStrategy } from "./FastStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
/**
 * @hidden
 */
export let FastStochasticOscillatorIndicator = /*@__PURE__*/ (() => {
    class FastStochasticOscillatorIndicator extends StrategyBasedIndicator {
        bj() {
            return new FastStochasticOscillatorIndicator();
        }
        constructor() {
            super();
            this.trendLineBrush = ((() => {
                let $ret = new Brush();
                $ret.fill = "#565656";
                return $ret;
            })());
            this.trendLineType = 10;
            this.trendLinePeriod = 3;
        }
        get_abv() {
            return new FastStochasticOscillatorIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return FastStochasticOscillatorIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(FastStochasticOscillatorIndicator.ach));
        }
        set acg(a) {
            this.h(FastStochasticOscillatorIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    FastStochasticOscillatorIndicator.$t = /*@__PURE__*/ markType(FastStochasticOscillatorIndicator, 'FastStochasticOscillatorIndicator', StrategyBasedIndicator.$);
    FastStochasticOscillatorIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(14, FastStochasticOscillatorIndicator.$);
    return FastStochasticOscillatorIndicator;
})();
