// igniteui-react-charts/esm2015/lib/FinancialIndicatorType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available types of marker than can be displayed by default by series with markers.
 */
export var FinancialIndicatorType = /*@__PURE__*/ (function (FinancialIndicatorType) {
    /**
     * Absolute volume oscillator indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["AbsoluteVolumeOscillator"] = 0] = "AbsoluteVolumeOscillator";
    /**
     * Accumulation distribution indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["AccumulationDistribution"] = 1] = "AccumulationDistribution";
    /**
     * Average directional index indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["AverageDirectionalIndex"] = 2] = "AverageDirectionalIndex";
    /**
     * Average true range indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["AverageTrueRange"] = 3] = "AverageTrueRange";
    /**
     * Bollinger band width indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["BollingerBandWidth"] = 4] = "BollingerBandWidth";
    /**
     * Chaikin oscillator indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["ChaikinOscillator"] = 5] = "ChaikinOscillator";
    /**
     * Chaikin volatility indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["ChaikinVolatility"] = 6] = "ChaikinVolatility";
    /**
     * Commodity channel index indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["CommodityChannelIndex"] = 7] = "CommodityChannelIndex";
    /**
     * Detrended price oscillator indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["DetrendedPriceOscillator"] = 8] = "DetrendedPriceOscillator";
    /**
     * Ease of movement indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["EaseOfMovement"] = 9] = "EaseOfMovement";
    /**
     * Fast stochastic oscillator indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["FastStochasticOscillator"] = 10] = "FastStochasticOscillator";
    /**
     * Force index indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["ForceIndex"] = 11] = "ForceIndex";
    /**
     * Full stochastic oscillator indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["FullStochasticOscillator"] = 12] = "FullStochasticOscillator";
    /**
     * Market facilitation index indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["MarketFacilitationIndex"] = 13] = "MarketFacilitationIndex";
    /**
     * Mass index indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["MassIndex"] = 14] = "MassIndex";
    /**
     * Median price indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["MedianPrice"] = 15] = "MedianPrice";
    /**
     * Money flow index indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["MoneyFlowIndex"] = 16] = "MoneyFlowIndex";
    /**
     * Moving average convergence divergence (MACD) indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["MovingAverageConvergenceDivergence"] = 17] = "MovingAverageConvergenceDivergence";
    /**
     * Negative volume index indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["NegativeVolumeIndex"] = 18] = "NegativeVolumeIndex";
    /**
     * On balance volume indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["OnBalanceVolume"] = 19] = "OnBalanceVolume";
    /**
     * Percentage price oscillator indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["PercentagePriceOscillator"] = 20] = "PercentagePriceOscillator";
    /**
     * Percentage volume oscillator indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["PercentageVolumeOscillator"] = 21] = "PercentageVolumeOscillator";
    /**
     * Positive volume index indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["PositiveVolumeIndex"] = 22] = "PositiveVolumeIndex";
    /**
     * Price volume trend indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["PriceVolumeTrend"] = 23] = "PriceVolumeTrend";
    /**
     * Rate of change and momentum indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["RateOfChangeAndMomentum"] = 24] = "RateOfChangeAndMomentum";
    /**
     * Relative strength index indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["RelativeStrengthIndex"] = 25] = "RelativeStrengthIndex";
    /**
     * Slow stochastic oscillator indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["SlowStochasticOscillator"] = 26] = "SlowStochasticOscillator";
    /**
     * Standard deviation indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["StandardDeviation"] = 27] = "StandardDeviation";
    /**
     * StochRSI indicator.
     * Stochastic Relative Strength Index.
     */
    FinancialIndicatorType[FinancialIndicatorType["StochRSI"] = 28] = "StochRSI";
    /**
     * TRIX indicator.
     * TRIple eXponential.
     */
    FinancialIndicatorType[FinancialIndicatorType["TRIX"] = 29] = "TRIX";
    /**
     * Typical price indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["TypicalPrice"] = 30] = "TypicalPrice";
    /**
     * Ultimate oscillator indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["UltimateOscillator"] = 31] = "UltimateOscillator";
    /**
     * Weighted close indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["WeightedClose"] = 32] = "WeightedClose";
    /**
     * Williams %R indicator.
     */
    FinancialIndicatorType[FinancialIndicatorType["WilliamsPercentR"] = 33] = "WilliamsPercentR";
    return FinancialIndicatorType;
})({});
/**
 * @hidden
 */
export let FinancialIndicatorType_$type = /*@__PURE__*/ markEnum('FinancialIndicatorType', 'AbsoluteVolumeOscillator,0|AccumulationDistribution,1|AverageDirectionalIndex,2|AverageTrueRange,3|BollingerBandWidth,4|ChaikinOscillator,5|ChaikinVolatility,6|CommodityChannelIndex,7|DetrendedPriceOscillator,8|EaseOfMovement,9|FastStochasticOscillator,10|ForceIndex,11|FullStochasticOscillator,12|MarketFacilitationIndex,13|MassIndex,14|MedianPrice,15|MoneyFlowIndex,16|MovingAverageConvergenceDivergence,17|NegativeVolumeIndex,18|OnBalanceVolume,19|PercentagePriceOscillator,20|PercentageVolumeOscillator,21|PositiveVolumeIndex,22|PriceVolumeTrend,23|RateOfChangeAndMomentum,24|RelativeStrengthIndex,25|SlowStochasticOscillator,26|StandardDeviation,27|StochRSI,28|TRIX,29|TypicalPrice,30|UltimateOscillator,31|WeightedClose,32|WilliamsPercentR,33');
