// igniteui-react-charts/esm2015/lib/FinancialChartZoomSliderType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * An enum representing the available series types to display in the zoom slider pane.
 */
export var FinancialChartZoomSliderType = /*@__PURE__*/ (function (FinancialChartZoomSliderType) {
    /**
     * Do not display the zoom slider pane.
     */
    FinancialChartZoomSliderType[FinancialChartZoomSliderType["None"] = 0] = "None";
    /**
     * In the zoom slider pane, match the series type in the price pane.
     */
    FinancialChartZoomSliderType[FinancialChartZoomSliderType["Auto"] = 1] = "Auto";
    /**
     * Display financial bar series in the zoom slider pane.
     */
    FinancialChartZoomSliderType[FinancialChartZoomSliderType["Bar"] = 2] = "Bar";
    /**
     * Display candle series in the zoom slider pane.
     */
    FinancialChartZoomSliderType[FinancialChartZoomSliderType["Candle"] = 3] = "Candle";
    /**
     * Display column series in the zoom slider pane.
     */
    FinancialChartZoomSliderType[FinancialChartZoomSliderType["Column"] = 4] = "Column";
    /**
     * Display line series in the zoom slider pane.
     */
    FinancialChartZoomSliderType[FinancialChartZoomSliderType["Line"] = 5] = "Line";
    /**
     * Display an area series in the zoom slider pane.
     */
    FinancialChartZoomSliderType[FinancialChartZoomSliderType["Area"] = 6] = "Area";
    return FinancialChartZoomSliderType;
})({});
/**
 * @hidden
 */
export let FinancialChartZoomSliderType_$type = /*@__PURE__*/ markEnum('FinancialChartZoomSliderType', 'None,0|Auto,1|Bar,2|Candle,3|Column,4|Line,5|Area,6');
