// igniteui-react-charts/esm2015/lib/FinancialValueList.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, IList$1_$type, Number_$type, markType } from "igniteui-react-core";
import { NotImplementedException } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
export let FinancialValueList = /*@__PURE__*/ (() => {
    class FinancialValueList extends Base {
        constructor(a, b, c, d, e) {
            super();
            this.d = null;
            this.b = null;
            this.a = null;
            this.c = null;
            this.e = null;
            this.d = a;
            this.b = b;
            this.a = d;
            this.c = c;
            this.e = e;
        }
        indexOf(a) {
            throw new NotImplementedException(0);
        }
        insert(a, b) {
            throw new NotImplementedException(0);
        }
        removeAt(a) {
            throw new NotImplementedException(0);
        }
        item(a, b) {
            if (arguments.length === 2) {
                return b;
            }
            else {
                let b = NaN;
                let c = NaN;
                let d = NaN;
                let e = NaN;
                let f = NaN;
                if (this.d != null && a >= 0 && a < this.d.count) {
                    b = this.d.item(a);
                }
                if (this.b != null && a >= 0 && a < this.b.count) {
                    c = this.b.item(a);
                }
                if (this.c != null && a >= 0 && a < this.c.count) {
                    d = this.c.item(a);
                }
                if (this.a != null && a >= 0 && a < this.a.count) {
                    e = this.a.item(a);
                }
                if (this.e != null && a >= 0 && a < this.e.count) {
                    f = this.e.item(a);
                }
                let g = isNaN_(b);
                let h = isNaN_(c);
                let i = isNaN_(d);
                let j = isNaN_(e);
                let k = isNaN_(f);
                if (!h && !i && !j) {
                    return (c + d + e) / 3;
                }
                if (!h && !i) {
                    return (c + d) / 2;
                }
                if (!g && !j) {
                    return (b + e) / 2;
                }
                if (!g) {
                    return b;
                }
                if (!h) {
                    return c;
                }
                if (!i) {
                    return d;
                }
                if (!j) {
                    return e;
                }
                return NaN;
            }
        }
        add(a) {
            throw new NotImplementedException(0);
        }
        clear() {
            throw new NotImplementedException(0);
        }
        contains(a) {
            throw new NotImplementedException(0);
        }
        copyTo(a, b) {
            throw new NotImplementedException(0);
        }
        get count() {
            let a = 0;
            let b = 0;
            let c = 0;
            let d = 0;
            let e = 0;
            if (this.d != null) {
                a = this.d.count;
            }
            if (this.b != null) {
                b = this.b.count;
            }
            if (this.c != null) {
                c = this.c.count;
            }
            if (this.a != null) {
                d = this.a.count;
            }
            if (this.e != null) {
                e = this.e.count;
            }
            let f = 0;
            f = Math.max(f, a);
            f = Math.max(f, b);
            f = Math.max(f, c);
            f = Math.max(f, d);
            f = Math.max(f, e);
            return f;
        }
        get isReadOnly() {
            return true;
        }
        remove(a) {
            throw new NotImplementedException(0);
        }
        getEnumerator() {
            throw new NotImplementedException(0);
        }
        getEnumeratorObject() {
            throw new NotImplementedException(0);
        }
    }
    FinancialValueList.$t = /*@__PURE__*/ markType(FinancialValueList, 'FinancialValueList', Base.$, [/*@__PURE__*/ IList$1_$type.specialize(Number_$type)]);
    return FinancialValueList;
})();
