// igniteui-react-charts/esm2015/lib/FinancialChartVolumeType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * An enum representing the available series types to display in the volume pane.
 */
export var FinancialChartVolumeType = /*@__PURE__*/ (function (FinancialChartVolumeType) {
    /**
     * Do not display the volume pane.
     */
    FinancialChartVolumeType[FinancialChartVolumeType["None"] = 0] = "None";
    /**
     * Display column series in the volume pane.
     */
    FinancialChartVolumeType[FinancialChartVolumeType["Column"] = 1] = "Column";
    /**
     * Display line series in the volume pane.
     */
    FinancialChartVolumeType[FinancialChartVolumeType["Line"] = 2] = "Line";
    /**
     * Display area series in the volume pane.
     */
    FinancialChartVolumeType[FinancialChartVolumeType["Area"] = 3] = "Area";
    return FinancialChartVolumeType;
})({});
/**
 * @hidden
 */
export let FinancialChartVolumeType_$type = /*@__PURE__*/ markEnum('FinancialChartVolumeType', 'None,0|Column,1|Line,2|Area,3');
