// igniteui-react-charts/esm2015/lib/FinancialChartIndicatorMenuItem.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, INotifyPropertyChanged_$type, PropertyChangedEventArgs, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let FinancialChartIndicatorMenuItem = /*@__PURE__*/ (() => {
    class FinancialChartIndicatorMenuItem extends Base {
        constructor() {
            super(...arguments);
            this.d = null;
            this.c = 0;
            this.f = null;
            this.g = null;
            this.a = false;
            this.propertyChanged = null;
        }
        get b() {
            return this.a;
        }
        set b(a) {
            let b = this.b;
            if (a != b) {
                this.a = a;
                if (this.propertyChanged != null) {
                    this.propertyChanged(this, new PropertyChangedEventArgs("IsChecked"));
                }
            }
        }
    }
    FinancialChartIndicatorMenuItem.$t = /*@__PURE__*/ markType(FinancialChartIndicatorMenuItem, 'FinancialChartIndicatorMenuItem', Base.$, [INotifyPropertyChanged_$type]);
    return FinancialChartIndicatorMenuItem;
})();
