// igniteui-react-charts/esm2015/lib/FinancialOverlay.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FinancialSeries } from "./FinancialSeries";
import { DependencyProperty } from "igniteui-react-core";
import { ISortingAxis_$type } from "./ISortingAxis";
import { typeCast, markType, markDep } from "igniteui-react-core";
import { Rect } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { AxisRange } from "./AxisRange";
import { PropertyMetadata } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
export let FinancialOverlay = /*@__PURE__*/ (() => {
    class FinancialOverlay extends FinancialSeries {
        get_e8() {
            return true;
        }
        constructor() {
            super();
            this.aa0 = false;
            this.aa0 = false;
        }
        pb(a, b, c, d) {
            super.pb(a, b, c, d);
            if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
                this.w6.i$i();
            }
            this.xn.b0.e(this.id);
            this.aa1();
            if (this.xp != null) {
                this.xp.c1();
            }
            this.rr(true);
        }
        get_hd() {
            return true;
        }
        q9(a, b, c, d) {
            switch (b) {
                case "FastItemsSource":
                    this.aa0 = false;
                    break;
                case FinancialOverlay.$$p[0]:
                    this.aa0 = false;
                    this.rr(false);
                    this.ql();
                    break;
                case "XAxis":
                    this.aa0 = false;
                    break;
            }
            super.q9(a, b, c, d);
        }
        b2() {
            if (this.b3 == 0) {
                return 1;
            }
            else {
                return super.b2();
            }
        }
        gw(a, b) {
            if (this.gv(a, b)) {
                return true;
            }
            return false;
        }
        ac() {
            super.ac();
            this.rr(false);
        }
        gd(a) {
            let b = this.dn != null ? this.dn.indexOf(a) : -1;
            let c = this.cv.by;
            c = c.copy();
            let d = this.cv.bx;
            let e = new Rect(0, 0, 0, 1, 1);
            let f = this.wl(this.cv);
            let g = new ScalerParams(0, e, e, this.w6.ch, f);
            let h = new ScalerParams(0, e, e, this.xp.ch, f);
            h.b = this.iy();
            if (b >= 0 && Rect.l_op_Inequality(c, null) && Rect.l_op_Inequality(d, null)) {
                if (this.w6 != null) {
                    let i = this.w6.el(b, g);
                    if (i < c.left + 0.1 * c.width) {
                        i = i + 0.4 * c.width;
                    }
                    if (i > c.right - 0.1 * c.width) {
                        i = i - 0.4 * c.width;
                    }
                    c.x = i - 0.5 * c.width;
                }
                if (this.xp != null && this.x1 != null && this.x0 != null && b < this.x1.count && b < this.x0.count) {
                    let j = this.xp.el(this.x1.item(b), h);
                    let k = this.xp.el(this.x0.item(b), h);
                    if (!isNaN_(j) && !isNaN_(k)) {
                        let l = Math.abs(j - k);
                        if (c.height < l) {
                            c.height = l;
                            c.y = Math.min(j, k);
                        }
                        else {
                            if (j < c.top + 0.1 * c.height) {
                                j = j + 0.4 * c.height;
                            }
                            if (j > c.bottom - 0.1 * c.height) {
                                j = j - 0.4 * c.height;
                            }
                            c.y = j - 0.5 * c.height;
                        }
                    }
                }
                this.c3.au(this.db, c, true);
            }
            return b >= 0;
        }
        ay(a) {
            if (a != null && a == this.xp && this.x1 != null && this.x0 != null) {
                return new AxisRange(this.x1.minimum, this.x0.maximum);
            }
            return null;
        }
    }
    FinancialOverlay.$t = /*@__PURE__*/ markType(FinancialOverlay, 'FinancialOverlay', FinancialSeries.$);
    FinancialOverlay.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, FinancialOverlay, 'raisePropertyChanged', ['IgnoreFirst:aa2:aa4', [1, 0]]);
    return FinancialOverlay;
})();
