// igniteui-react-charts/esm2015/lib/FinancialChartEnumCollectionPickerVisualData.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { FinancialChartVisualData } from "./FinancialChartVisualData";
/**
 * @hidden
 */
export let FinancialChartEnumCollectionPickerVisualData = /*@__PURE__*/ (() => {
    class FinancialChartEnumCollectionPickerVisualData extends Base {
        constructor() {
            super(...arguments);
            this.a = null;
            this.b = null;
            this.k = 0;
            this.l = 0;
            this.m = 0;
            this.j = 0;
            this.f = 0;
            this.g = 0;
            this.h = 0;
            this.e = 0;
            this.n = 0;
            this.i = 0;
            this.d = false;
            this.c = false;
        }
        o() {
            return FinancialChartVisualData.y(["items", "selectedItems", "pickerLeft", "pickerTop", "pickerWidth", "pickerHeight", "buttonLeft", "buttonTop", "buttonWidth", "buttonHeight", "width", "height", "isVisible", "isPickerVisible"], [this.a, this.b, this.k, this.l, this.m, this.j, this.f, this.g, this.h, this.e, this.n, this.i, this.d, this.c]);
        }
    }
    FinancialChartEnumCollectionPickerVisualData.$t = /*@__PURE__*/ markType(FinancialChartEnumCollectionPickerVisualData, 'FinancialChartEnumCollectionPickerVisualData');
    return FinancialChartEnumCollectionPickerVisualData;
})();
