// igniteui-react-charts/esm2015/lib/FinancialChartTypePicker.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FinancialChartDropDownControl } from "./FinancialChartDropDownControl";
import { FinancialChartTypePickerViewModel } from "./FinancialChartTypePickerViewModel";
import { FinancialChartTypePickerView } from "./FinancialChartTypePickerView";
import { runOn, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let FinancialChartTypePicker = /*@__PURE__*/ (() => {
    class FinancialChartTypePicker extends FinancialChartDropDownControl {
        constructor(a) {
            super("financialChartTypePicker", a, FinancialChartTypePicker.z);
            this.o = new FinancialChartTypePickerView();
            this.p = new FinancialChartTypePickerViewModel();
            this.p.onSelectionChanged = runOn(this, this.v);
            this.p.b = runOn(this.o, this.o.j);
            this.o.a = this.p;
        }
        get w() {
            return this.y.chartType;
        }
        set w(a) {
            this.y.chartType = a;
        }
        get x() {
            return this.o;
        }
        get y() {
            return this.p;
        }
        get_s() {
            return ".financialChartTypePickerDropDownButton";
        }
        get s() {
            return this.get_s();
        }
        get_t() {
            return ".financialChartTypePickerDropDownMenu";
        }
        get t() {
            return this.get_t();
        }
    }
    FinancialChartTypePicker.$t = /*@__PURE__*/ markType(FinancialChartTypePicker, 'FinancialChartTypePicker', FinancialChartDropDownControl.$);
    FinancialChartTypePicker.z = "<button class='financialChartTypePickerDropDownButton' data-value='Auto'>\r\n\t\t<span class='ui-icon ui-icon-changeable'></span>\r\n\t\t<span class='ui-icon-toolbar-dd'></span>\r\n\t</button>\r\n\r\n\t<div class='financialChartTypePickerDropDownMenu'>\r\n\t\t<ul>\r\n\t\t\t<li class='financialChartListItem' data-value='Auto'>\r\n\t\t\t<span class='ui-icon ui-icon-chartAuto'></span>\r\n\t\t\t<span>Auto</span>\r\n\t\t\t</li>\r\n\t\t\t<li class='financialChartListItem' data-value='Bar'>\r\n\t\t\t\t<span class='ui-icon ui-icon-BarIcon'></span>\r\n\t\t\t\t<span>Bar</span>\r\n\t\t\t</li>\r\n\t\t\t<li class='financialChartListItem' data-value='Line'>\r\n\t\t\t\t<span class='ui-icon ui-icon-LineIcon'></span>\r\n\t\t\t\t<span>Line</span>\r\n\t\t\t</li>\r\n\t\t\t<li class='financialChartListItem' data-value='Candle'>\r\n\t\t\t\t<span class='ui-icon ui-icon-CandleIcon'></span>\r\n\t\t\t\t<span>Candle</span>\r\n\t\t\t</li>\r\n\t\t\t<li class='financialChartListItem' data-value='Column'>\r\n\t\t\t\t<span class='ui-icon ui-icon-ColumnIcon'></span>\r\n\t\t\t\t<span>Column</span>\r\n\t\t\t</li>\r\n\t\t</ul>\r\n\t</div>";
    return FinancialChartTypePicker;
})();
