// igniteui-react-charts/esm2015/lib/FinancialEventArgs_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { EventArgs, Base, markType, typeCast, runOn, delegateCombine, enumGetBox, Point_$type, String_$type, Number_$type, fromEnum, toEnum, Date_$type, markDep, Delegate_$type } from "igniteui-react-core";
import { IBucketizer_$type } from "./IBucketizer";
import { ArgumentNullException } from "igniteui-react-core";
import { IFastItemColumn$1_$type } from "igniteui-react-core";
import { Series } from "./Series";
import { Rect } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { ISortingAxis_$type } from "./ISortingAxis";
import { SeriesView } from "./SeriesView";
import { truncate, intDivide, isInfinity, isNaN_ } from "igniteui-react-core";
import { IHasCategoryAxis_$type } from "./IHasCategoryAxis";
import { IHasCategoryModePreference_$type } from "./IHasCategoryModePreference";
import { Brush } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { NumericYAxis } from "./NumericYAxis";
import { HighlightingInfo } from "igniteui-react-core";
import { CategoryTransitionInMode_$type } from "./CategoryTransitionInMode";
import { CategoryTransitionSourceFramePreparer } from "./CategoryTransitionSourceFramePreparer";
import { CategoryFrame } from "./CategoryFrame";
import { CategoryMode_$type } from "./CategoryMode";
import { ITimeRangeSelectorAxis_$type } from "./ITimeRangeSelectorAxis";
import { FastReflectionHelper } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { SeriesLayerPropertyOverlay } from "./SeriesLayerPropertyOverlay";
import { Dictionary$2 } from "igniteui-react-core";
import { CategorySeriesRenderManager } from "./CategorySeriesRenderManager";
import { IFastItemsSource_$type } from "igniteui-react-core";
import { FinancialValueList } from "./FinancialValueList";
import { CategoryXAxis } from "./CategoryXAxis";
import { SafeSortedReadOnlyDoubleCollection } from "./SafeSortedReadOnlyDoubleCollection";
import { SafeReadOnlyDoubleCollection } from "./SafeReadOnlyDoubleCollection";
import { SafeEnumerable } from "./SafeEnumerable";
import { CalculatedColumn } from "./CalculatedColumn";
import { TrendCalculators } from "igniteui-react-core";
import { ColumnSupportingCalculation } from "./ColumnSupportingCalculation";
import { IDateTimeAxis_$type } from "./IDateTimeAxis";
import { PropertyMetadata } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
import { dateToStringFormat } from "igniteui-react-core";
import { SupportingCalculation$1 } from "./SupportingCalculation$1";
import { SRProvider } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
export let FinancialEventArgs = /*@__PURE__*/ (() => {
    class FinancialEventArgs extends EventArgs {
        constructor(a, b, c, d, e) {
            super();
            this._position = 0;
            this._count = 0;
            this._dataSource = null;
            this._supportingCalculations = null;
            this._basedOn = null;
            this.position = a;
            this.count = b;
            this.dataSource = c;
            this.supportingCalculations = d;
            if (c != null) {
                c.series = e;
            }
        }
        get position() {
            return this._position;
        }
        set position(a) {
            this._position = a;
        }
        get count() {
            return this._count;
        }
        set count(a) {
            this._count = a;
        }
        get dataSource() {
            return this._dataSource;
        }
        set dataSource(a) {
            this._dataSource = a;
        }
        get supportingCalculations() {
            return this._supportingCalculations;
        }
        set supportingCalculations(a) {
            this._supportingCalculations = a;
        }
        get basedOn() {
            return this._basedOn;
        }
        set basedOn(a) {
            this._basedOn = a;
        }
    }
    FinancialEventArgs.$t = /*@__PURE__*/ markType(FinancialEventArgs, 'FinancialEventArgs', EventArgs.$);
    return FinancialEventArgs;
})();
/**
 * @hidden
 */
export let FinancialBucketCalculator = /*@__PURE__*/ (() => {
    class FinancialBucketCalculator extends Base {
        constructor(a) {
            super();
            this.a = null;
            this.c = 0;
            this.d = 0;
            this.b = 0;
            if (a == null) {
                throw new ArgumentNullException(0, "view");
            }
            this.a = a;
            this.c = -1;
            this.d = this.d;
            this.b = 0;
        }
        getBucket(a) {
            return null;
        }
        getBucketWithoutUnknowns(a) {
            return this.getBucket(a);
        }
        getErrorBucket(a, b) {
            return NaN;
        }
        getBucketInfo(a, b, c, d) {
            a = this.c;
            b = this.d;
            c = this.b;
            d = this.a.b2.id;
            return {
                p0: a,
                p1: b,
                p2: c,
                p3: d
            };
        }
        e(a) {
            let b = this.a.by;
            let c = this.a.bx;
            let d = this.a.bx;
            if (this.a.b2 != null) {
                d = this.a.b2.getEffectiveViewport1(this.a);
            }
            if (b.isEmpty || c.isEmpty || this.a.b2.w6 == null) {
                this.b = 0;
                return;
            }
            let e = (this.a.b2.w6 != null) ? this.a.b2.w6.ch : false;
            let f = new ScalerParams(0, b, c, e, d);
            let g = this.a.b2.wj(this.a);
            let h = typeCast(ISortingAxis_$type, this.a.b2.w6);
            if (h == null || h.i$b == null) {
                let i = Math.floor(this.a.b2.w6.eo(g.left, f));
                let j = Math.ceil(this.a.b2.w6.eo(g.right, f));
                if (this.a.b2.w6.ch) {
                    j = Math.ceil(this.a.b2.w6.eo(g.left, f));
                    i = Math.floor(this.a.b2.w6.eo(g.right, f));
                }
                let k = Math.floor((j - i + 1) * a / g.width);
                this.b = truncate(Math.max(1, k));
                this.c = truncate(Math.floor(i / this.b));
                this.d = truncate(Math.ceil(j / this.b));
            }
            else {
                this.c = h.i$e(b, c, d, g);
                this.d = h.i$g(b, c, d, g);
                this.b = 1;
            }
        }
        cacheValues() {
        }
        unCacheValues() {
        }
    }
    FinancialBucketCalculator.$t = /*@__PURE__*/ markType(FinancialBucketCalculator, 'FinancialBucketCalculator', Base.$, [IBucketizer_$type]);
    return FinancialBucketCalculator;
})();
/**
 * @hidden
 */
export let FinancialSeries = /*@__PURE__*/ (() => {
    class FinancialSeries extends Series {
        ht() {
            let a = super.ht();
            {
                let b = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.ad = "OpenMemberPath";
                    $ret.p = true;
                    $ret.y = "HighlightedOpenMemberPath";
                    $ret.q = true;
                    return $ret;
                })());
                b.valueResolving = delegateCombine(b.valueResolving, runOn(this, this.z9));
                a.add(b);
            }
            {
                let c = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.ad = "HighMemberPath";
                    $ret.p = true;
                    $ret.y = "HighlightedHighMemberPath";
                    $ret.q = true;
                    return $ret;
                })());
                c.valueResolving = delegateCombine(c.valueResolving, runOn(this, this.z7));
                a.add(c);
            }
            {
                let d = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.ad = "LowMemberPath";
                    $ret.p = true;
                    $ret.y = "HighlightedLowMemberPath";
                    $ret.q = true;
                    return $ret;
                })());
                d.valueResolving = delegateCombine(d.valueResolving, runOn(this, this.z8));
                a.add(d);
            }
            {
                let e = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.ad = "CloseMemberPath";
                    $ret.p = true;
                    $ret.y = "HighlightedCloseMemberPath";
                    $ret.q = true;
                    return $ret;
                })());
                e.valueResolving = delegateCombine(e.valueResolving, runOn(this, this.z6));
                a.add(e);
            }
            {
                let f = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.ad = "VolumeMemberPath";
                    $ret.p = true;
                    $ret.y = "HighlightedVolumeMemberPath";
                    $ret.q = true;
                    return $ret;
                })());
                f.valueResolving = delegateCombine(f.valueResolving, runOn(this, this.aaa));
                a.add(f);
            }
            {
                let g = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.r = true;
                    $ret.ad = "Brush";
                    $ret.p = true;
                    $ret.y = "ActualBrush";
                    $ret.q = true;
                    return $ret;
                })());
                g.valueResolving = delegateCombine(g.valueResolving, runOn(this, this.r0));
                a.add(g);
            }
            {
                let h = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.r = true;
                    $ret.ad = "Outline";
                    $ret.p = true;
                    $ret.y = "ActualOutline";
                    $ret.q = true;
                    return $ret;
                })());
                h.valueResolving = delegateCombine(h.valueResolving, runOn(this, this.r0));
                a.add(h);
            }
            {
                let i = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.r = true;
                    $ret.ad = "NegativeBrush";
                    $ret.p = true;
                    $ret.q = true;
                    return $ret;
                })());
                i.valueResolving = delegateCombine(i.valueResolving, runOn(this, this.r0));
                a.add(i);
            }
            {
                let j = ((() => {
                    let $ret = new SeriesLayerPropertyOverlay();
                    $ret.r = true;
                    $ret.ad = "NegativeOutline";
                    $ret.p = true;
                    $ret.q = true;
                    return $ret;
                })());
                j.valueResolving = delegateCombine(j.valueResolving, runOn(this, this.r0));
                a.add(j);
            }
            let k = ((() => {
                let $ret = new SeriesLayerPropertyOverlay();
                $ret.p = true;
                $ret.ad = "IsTransitionInEnabled";
                $ret.x = true;
                $ret.q = true;
                return $ret;
            })());
            a.add(k);
            let l = ((() => {
                let $ret = new SeriesLayerPropertyOverlay();
                $ret.p = true;
                $ret.ad = "TransitionInMode";
                $ret.x = enumGetBox(CategoryTransitionInMode_$type, 20);
                $ret.q = true;
                return $ret;
            })());
            a.add(l);
            return a;
        }
        z9(a, b) {
            if (this.zj != null) {
                b.value = this.zj;
                return;
            }
            b.value = this.zw;
        }
        z7(a, b) {
            if (this.zf != null) {
                b.value = this.zf;
                return;
            }
            b.value = this.zn;
        }
        z8(a, b) {
            if (this.zh != null) {
                b.value = this.zh;
                return;
            }
            b.value = this.zs;
        }
        z6(a, b) {
            if (this.zd != null) {
                b.value = this.zd;
                return;
            }
            b.value = this.za;
        }
        aaa(a, b) {
            if (this.zl != null) {
                b.value = this.zl;
                return;
            }
            b.value = this.z0;
        }
        kv() {
            if (this.d3()) {
                return this.bn.xc;
            }
            else {
                return this.bn.w8;
            }
        }
        q1(a) {
            super.q1(a);
            this.xn = a;
        }
        get_e6() {
            return true;
        }
        ai() {
            return [this.w6, this.xp];
        }
        kq(a, b) {
            let c = this.nd(b);
            if (c == FinancialSeries.$$p[11]) {
                return this.kr(a, b, this.zw);
            }
            if (c == FinancialSeries.$$p[1]) {
                return this.kr(a, b, this.zn);
            }
            if (c == FinancialSeries.$$p[9]) {
                return this.kr(a, b, this.zs);
            }
            if (c == FinancialSeries.$$p[0]) {
                return this.kr(a, b, this.za);
            }
            if (c == FinancialSeries.$$p[13]) {
                return this.kr(a, b, this.z0);
            }
            if (c == FinancialSeries.$$p[5]) {
                return this.kr(a, b, this.zj);
            }
            if (c == FinancialSeries.$$p[3]) {
                return this.kr(a, b, this.zf);
            }
            if (c == FinancialSeries.$$p[4]) {
                return this.kr(a, b, this.zh);
            }
            if (c == FinancialSeries.$$p[2]) {
                return this.kr(a, b, this.zd);
            }
            if (c == FinancialSeries.$$p[6]) {
                return this.kr(a, b, this.zl);
            }
            return super.kq(a, b);
        }
        md(a) {
            let b = this.nd(a);
            if (b == FinancialSeries.$$p[11]) {
                return this.zw;
            }
            if (b == FinancialSeries.$$p[1]) {
                return this.zn;
            }
            if (b == FinancialSeries.$$p[9]) {
                return this.zs;
            }
            if (b == FinancialSeries.$$p[0]) {
                return this.za;
            }
            if (b == FinancialSeries.$$p[13]) {
                return this.z0;
            }
            if (b == FinancialSeries.$$p[5]) {
                return this.zj;
            }
            if (b == FinancialSeries.$$p[3]) {
                return this.zf;
            }
            if (b == FinancialSeries.$$p[4]) {
                return this.zh;
            }
            if (b == FinancialSeries.$$p[2]) {
                return this.zd;
            }
            if (b == FinancialSeries.$$p[6]) {
                return this.zl;
            }
            return super.md(a);
        }
        y6() {
            return 2;
        }
        y4() {
            return 4;
        }
        aj(a) {
            let b = this.an(this.w6, this.w8.f, this.y3(this.cv), this.y5(this.cv), this.wh(a), true);
            if (b == null) {
                return null;
            }
            let c = b[0];
            let d = b[1];
            let e = new Array(1);
            let f = new Array(2);
            let g = this.y6();
            f[0] = { $type: Point_$type, x: c[0], y: c[g] };
            f[1] = { $type: Point_$type, x: d[0], y: d[g] };
            e[0] = f;
            return e;
        }
        al(a) {
            let b = this.an(this.w6, this.w8.f, this.y3(this.cv), this.y5(this.cv), this.wh(a), true);
            if (b == null) {
                return null;
            }
            let c = b[0];
            let d = b[1];
            let e = new Array(1);
            let f = new Array(2);
            f[0] = { $type: Point_$type, x: c[0], y: c[1] };
            f[1] = { $type: Point_$type, x: d[0], y: d[1] };
            e[0] = f;
            return e;
        }
        constructor() {
            super();
            this.xx = null;
            this.xv = null;
            this.xw = null;
            this.xu = null;
            this.xy = null;
            this.x7 = false;
            this.assigningCategoryStyle = null;
            this.xf = null;
            this.yh = new Dictionary$2(IFastItemColumn$1_$type.specialize(Number_$type), String_$type, 0);
            this.yi = new Dictionary$2(String_$type, String_$type, 0);
            this.x6 = false;
            this.xt = null;
            this.xa = null;
            this.xc = null;
            this.w8 = null;
            this.xi = null;
            this.typical = null;
            this.typicalBasedOn = null;
            this.x5 = false;
            this.yu = null;
            this.ys = null;
            this.yt = null;
            this.yr = null;
            this.yv = null;
            this.xb = null;
            this.w7 = null;
            this.xd = 0;
            this.xo = null;
            this.xq = null;
            this.xf = new CategorySeriesRenderManager();
            this.x7 = false;
            this.xi = new CategoryTransitionSourceFramePreparer();
            this.xb = new CategoryFrame(3);
            this.ab = FinancialSeries.$;
            this.xe = 0;
        }
        vw() {
            return this.aaw;
        }
        vx() {
            return this.aaw;
        }
        o2() {
            super.o2();
            this.w6 = null;
            this.xp = null;
        }
        get_ek() {
            return true;
        }
        get_f5() {
            return this.xp != null && this.xp.ch;
        }
        get x2() {
            return this.xx;
        }
        set x2(a) {
            if (this.xx != a) {
                let b = this.xx;
                this.xx = a;
                this.raisePropertyChanged("OpenColumn", b, this.xx);
            }
        }
        get x0() {
            return this.xv;
        }
        set x0(a) {
            if (this.xv != a) {
                let b = this.xv;
                this.xv = a;
                this.raisePropertyChanged("HighColumn", b, this.xv);
            }
        }
        get x1() {
            return this.xw;
        }
        set x1(a) {
            if (this.xw != a) {
                let b = this.xw;
                this.xw = a;
                this.raisePropertyChanged("LowColumn", b, this.xw);
            }
        }
        get xz() {
            return this.xu;
        }
        set xz(a) {
            if (this.xu != a) {
                let b = this.xu;
                this.xu = a;
                this.raisePropertyChanged("CloseColumn", b, this.xu);
            }
        }
        get x4() {
            return this.xy;
        }
        set x4(a) {
            if (this.xy != a) {
                let b = this.xy;
                this.xy = a;
                this.raisePropertyChanged("VolumeColumn", b, this.xy);
            }
        }
        gb() {
            return this.highlightedItemsSource != null || (this.zj != null && this.zf != null && this.zh != null && this.zd != null);
        }
        ye() {
            return (this.assigningCategoryStyle != null && this.x7) || this.bx != 1 || this.cp != 1 || this.co != 1;
        }
        aad(a) {
            if (this.assigningCategoryStyle != null && this.x7) {
                this.assigningCategoryStyle(this, a);
            }
        }
        aab(a, b, c, d, e, f, g) {
            let h = this.bx != 1;
            let i = this.cp != 1;
            let j = this.co != 1;
            this.xf.a3(this.xf.b, d.cn, a, b, c, d, e, f, false, g);
            this.xf.ab(this.ib);
            let k = this.xf.ay;
            let l = null;
            let m = null;
            let n = null;
            let o = this.dn;
            if (h && this.db != null) {
                l = this.db.dl.getHighlightingInfo(this, o, k.z, k.y, false);
                k.c = l;
            }
            if (i && this.db != null) {
                m = this.db.dm.getHighlightingInfo(this, o, k.z, k.y, false);
                if (m == null && this.bn != null) {
                    m = this.db.dm.getHighlightingInfo(this.bn, this.bn.dn, k.z, k.y, false);
                }
                k.d = m;
            }
            if (j && this.db != null) {
                n = this.db.dk.getHighlightingInfo(this, o, k.z, k.y, false);
                k.b = n;
            }
            k.h = f;
            this.aad(k);
            this.xf.z(this, l, m, n, f, h, i, j, b < 0);
        }
        yw(a) {
            return this.it(a, this.w6);
        }
        ka(a, b) {
            return this.kb(a, b, this.w6, runOn(this, this.yw), new FinancialValueList(this.x2, this.x0, this.x1, this.xz, this.x4));
        }
        j8(a, b) {
            return this.j9(a, b, this.w6, runOn(this, this.yw), new FinancialValueList(this.x2, this.x0, this.x1, this.xz, this.x4));
        }
        ip(a, b, c, d, e) {
            if (c == null) {
                return Number.POSITIVE_INFINITY;
            }
            let f = this.w6.lj;
            return this.iq(a, b, this.w6, d, e, f, runOn(this, this.yw));
        }
        getOffsetValue() {
            return this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
        }
        getCategoryWidth() {
            return this.w6.getCategorySize(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
        }
        wc(a, b, c) {
            return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, null, null, null);
        }
        i5(a, b, c) {
            let d = this.getEffectiveViewport1(this.cv);
            let e = this.yx(this.cv.by, this.cv.bx, d);
            let f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
            return this.i7(new FinancialValueList(this.x2, this.x0, this.x1, this.xz, this.x4), a, this.w6, f, e, runOn(this, this.yw), b, c);
        }
        i1(a, b, c) {
            let d = this.getEffectiveViewport1(this.cv);
            let e = this.yx(this.cv.by, this.cv.bx, d);
            let f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
            return this.i7(this.x0, a, this.w6, f, e, runOn(this, this.yw), b, c);
        }
        i3(a, b, c) {
            let d = this.getEffectiveViewport1(this.cv);
            let e = this.yx(this.cv.by, this.cv.bx, d);
            let f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
            return this.i7(this.x1, a, this.w6, f, e, runOn(this, this.yw), b, c);
        }
        yy(a, b, c) {
            let d = this.getEffectiveViewport1(this.cv);
            let e = this.yx(this.cv.by, this.cv.bx, d);
            let f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
            return this.i7(this.xz, a, this.w6, f, e, runOn(this, this.yw), b, c);
        }
        yz(a, b, c) {
            let d = this.getEffectiveViewport1(this.cv);
            let e = this.yx(this.cv.by, this.cv.bx, d);
            let f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
            return this.i7(this.x2, a, this.w6, f, e, runOn(this, this.yw), b, c);
        }
        y0(a, b, c) {
            let d = this.getEffectiveViewport1(this.cv);
            let e = this.yx(this.cv.by, this.cv.bx, d);
            let f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
            return this.i7(this.x4, a, this.w6, f, e, runOn(this, this.yw), b, c);
        }
        v8(a, b, c) {
            return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, runOn(this, this.i1), (d, e) => this.kb(d, e, this.w6, runOn(this, this.yw), this.x0), (d, e) => this.j9(d, e, this.w6, runOn(this, this.yw), this.x0));
        }
        wa(a, b, c) {
            return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, runOn(this, this.i3), (d, e) => this.kb(d, e, this.w6, runOn(this, this.yw), this.x1), (d, e) => this.j9(d, e, this.w6, runOn(this, this.yw), this.x1));
        }
        aay(a, b, c) {
            return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, runOn(this, this.yz), (d, e) => this.kb(d, e, this.w6, runOn(this, this.yw), this.x2), (d, e) => this.j9(d, e, this.w6, runOn(this, this.yw), this.x2));
        }
        aax(a, b, c) {
            return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, runOn(this, this.yy), (d, e) => this.kb(d, e, this.w6, runOn(this, this.yw), this.xz), (d, e) => this.j9(d, e, this.w6, runOn(this, this.yw), this.xz));
        }
        aaz(a, b, c) {
            return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, runOn(this, this.y0), (d, e) => this.kb(d, e, this.w6, runOn(this, this.yw), this.x4), (d, e) => this.j9(d, e, this.w6, runOn(this, this.yw), this.x4));
        }
        qb() {
            super.qb();
            if (this.w6 != null) {
                this.w6.ib(false);
            }
            if (this.xp != null) {
                this.xp.ib(false);
            }
        }
        sr(a, b) {
            this.xn.b0.e(this.id);
            this.rr(false);
        }
        sq(a, b) {
            if (this.g1 && this.jq < 0.05) {
                this.g2 = true;
                this.g5 = false;
                this.gk = false;
            }
            this.xn.b0.e(this.id);
            this.rr(false);
        }
        x3(a, b, c) {
            let d = this.dk(b);
            this.yh.addItem(d, b);
            this.yi.addItem(b, c);
            return d;
        }
        z5(a, b) {
            if (b == null) {
                return;
            }
            a.deregisterColumn(b);
            let c = this.yh.item(b);
            this.yi.removeItem(c);
            this.yh.removeItem(b);
        }
        q9(a, b, c, d) {
            super.q9(a, b, c, d);
            switch (b) {
                case "SyncLink":
                    if (this.c3 != null && this.db != null) {
                        this.xn.b0.e(this.id);
                        this.rr(false);
                    }
                    break;
                case "SeriesViewer":
                    if (c != null && d == null) {
                        if (this.w6 != null) {
                            this.w6.b0(this);
                        }
                        if (this.xp != null) {
                            this.xp.b0(this);
                        }
                    }
                    if (c == null && d != null) {
                        if (this.w6 != null) {
                            this.w6.ct(this);
                        }
                        if (this.xp != null) {
                            this.xp.ct(this);
                        }
                    }
                    this.xn.b0.e(this.id);
                    this.rr(false);
                    break;
                case "FastItemsSource":
                    this.x6 = true;
                    if (typeCast(IFastItemsSource_$type, c) != null) {
                        this.z5(typeCast(IFastItemsSource_$type, c), this.x2);
                        this.z5(typeCast(IFastItemsSource_$type, c), this.x0);
                        this.z5(typeCast(IFastItemsSource_$type, c), this.x1);
                        this.z5(typeCast(IFastItemsSource_$type, c), this.xz);
                        this.z5(typeCast(IFastItemsSource_$type, c), this.x4);
                        this.x2 = null;
                        this.x0 = null;
                        this.x1 = null;
                        this.xz = null;
                        this.x4 = null;
                    }
                    if (typeCast(IFastItemsSource_$type, d) != null) {
                        if (stringIsNullOrEmpty(this.zw) == false) {
                            this.x2 = this.x3(typeCast(IFastItemsSource_$type, d), this.zw, "OpenColumn");
                        }
                        if (stringIsNullOrEmpty(this.zn) == false) {
                            this.x0 = this.x3(typeCast(IFastItemsSource_$type, d), this.zn, "HighColumn");
                        }
                        if (stringIsNullOrEmpty(this.zs) == false) {
                            this.x1 = this.x3(typeCast(IFastItemsSource_$type, d), this.zs, "LowColumn");
                        }
                        if (stringIsNullOrEmpty(this.za) == false) {
                            this.xz = this.x3(typeCast(IFastItemsSource_$type, d), this.za, "CloseColumn");
                        }
                        if (stringIsNullOrEmpty(this.z0) == false) {
                            this.x4 = this.x3(typeCast(IFastItemsSource_$type, d), this.z0, "VolumeColumn");
                        }
                    }
                    this.x6 = false;
                    if (this.xp != null && !this.xp.c1()) {
                        this.xn.b0.e(this.id);
                        this.rr(this.gg && c != null);
                    }
                    break;
                case FinancialSeries.$$p[11]:
                    if (this.dn != null) {
                        this.z5(this.dn, this.x2);
                        this.x2 = this.x3(this.dn, this.zw, "OpenColumn");
                    }
                    break;
                case "OpenColumn":
                    if (this.xp != null && !this.xp.c1() && !this.x6) {
                        this.xn.b0.e(this.id);
                        this.rr(this.gg && this.d7);
                    }
                    break;
                case FinancialSeries.$$p[1]:
                    if (this.dn != null) {
                        this.z5(this.dn, this.x0);
                        this.x0 = this.x3(this.dn, this.zn, "HighColumn");
                    }
                    break;
                case "HighColumn":
                    if (this.xp != null && !this.xp.c1() && !this.x6) {
                        this.xn.b0.e(this.id);
                        this.rr(this.gg && this.d7);
                    }
                    break;
                case FinancialSeries.$$p[9]:
                    if (this.dn != null) {
                        this.z5(this.dn, this.x1);
                        this.x1 = this.x3(this.dn, this.zs, "LowColumn");
                    }
                    break;
                case "LowColumn":
                    if (this.xp != null && !this.xp.c1() && !this.x6) {
                        this.xn.b0.e(this.id);
                        this.rr(this.gg && this.d7);
                    }
                    break;
                case FinancialSeries.$$p[0]:
                    if (this.dn != null) {
                        this.z5(this.dn, this.xz);
                        this.xz = this.x3(this.dn, this.za, "CloseColumn");
                    }
                    break;
                case "CloseColumn":
                    if (this.xp != null && !this.xp.c1() && !this.x6) {
                        this.xn.b0.e(this.id);
                        this.rr(this.gg && this.d7);
                    }
                    this.aaf();
                    break;
                case FinancialSeries.$$p[13]:
                    if (this.dn != null) {
                        this.z5(this.dn, this.x4);
                        this.x4 = this.x3(this.dn, this.z0, "VolumeColumn");
                    }
                    break;
                case "VolumeColumn":
                    if (this.xp != null && !this.xp.c1() && !this.x6) {
                        this.xn.b0.e(this.id);
                        this.rr(this.gg && this.d7);
                    }
                    break;
                case FinancialSeries.$$p[14]:
                    if (c != null) {
                        c.b0(this);
                    }
                    if (d != null) {
                        d.ct(this);
                    }
                    this.xn.b0.e(this.id);
                    this.rr(false);
                    this.ql();
                    this.xo = typeCast(ITimeRangeSelectorAxis_$type, this.w6);
                    break;
                case FinancialSeries.$$p[15]:
                    if (c != null) {
                        c.b0(this);
                    }
                    if (d != null) {
                        d.ct(this);
                    }
                    this.xn.b0.e(this.id);
                    if (this.xp != null) {
                        this.xp.c1();
                    }
                    {
                        this.rr(false);
                    }
                    this.ql();
                    break;
                case "TransitionProgress":
                    this.xc.c(this.jq, this.xa, this.w8);
                    if (this.d5(this.cv)) {
                        return;
                    }
                    if (this.jq == 1) {
                        this.w8.w();
                        this.aae(this.w8, this.xn);
                        if (this.g1) {
                            this.g1 = false;
                            this.o6();
                            this.dc.r = truncate(this.i9());
                            this.dc.b = this.dd;
                        }
                        if (this.g4) {
                            this.g4 = false;
                            this.o6();
                            this.dc.r = truncate(this.i9());
                            this.dc.b = this.dd;
                            this.q0();
                        }
                        this.qo();
                    }
                    else {
                        this.aae(this.xc, this.xn);
                    }
                    if (this.db != null) {
                        this.db.qo(this);
                    }
                    break;
                case FinancialSeries.$$p[10]:
                    this.rr(false);
                    break;
                case "IsCustomCategoryStyleAllowed":
                    this.x7 = this.ya();
                    this.rr(false);
                    break;
                case "CategoryMode":
                    if (this.w6 != null) {
                        this.w6.ma(true);
                    }
                    break;
            }
        }
        x9(a) {
            if (typeCast(NumericYAxis.$, a) !== null) {
                return true;
            }
            return false;
        }
        x8(a) {
            if (typeCast(CategoryXAxis.$, a) !== null || a.ce) {
                return true;
            }
            return false;
        }
        wz(a, b, c) {
            if (b < a || a < 0 || a > this.dn.count || b < 0 || b > this.dn.count || c == null) {
                return null;
            }
            let d = new Array((b - a) + 1);
            let e = c.cn;
            let f = null;
            if (e) {
                f = c.i$b;
            }
            for (let g = a; g <= b; g++) {
                let h = g;
                if (e) {
                    h = f._inner[h];
                }
                d[g - a] = this.dn.item(g);
            }
            return d;
        }
        y3(a) {
            return a.b0.b;
        }
        y5(a) {
            return a.b0.c;
        }
        wy(a, b) {
            return this.wz(a, b, this.xp);
        }
        get_fe() {
            return true;
        }
        o3() {
            super.o3();
            this.xt = null;
        }
        dg(a, b, c) {
            if (this.ed && this.fz(this.bx) && !c) {
                let d = this.xs(a, b, this.xt);
                this.xt = d;
                return d;
            }
            else {
                let e = this.xr(a, b, this.xt);
                this.xt = e;
                return e;
            }
        }
        y7(a) {
            let b = this.w9();
            let c = this.cv.bx;
            let d = this.cv.by;
            let e = this.w6;
            let f = 0;
            if (e.cp) {
                let g = (a.y - d.top) / d.height;
                f = g * c.height + c.top;
            }
            else {
                let h = (a.x - d.left) / d.width;
                f = h * c.width + c.left;
            }
            let i = b.f.count;
            let j = 0;
            for (j = 0; j < i; j++) {
                let k = b.f._inner[j];
                let l = null;
                if (j < i - 1) {
                    l = b.f._inner[j + 1];
                }
                if (k[0] <= f && l == null || l[0] >= f) {
                    if (l != null) {
                        if (Math.abs(k[0] - f) < Math.abs(l[0] - f)) {
                            return j;
                        }
                        else {
                            return j + 1;
                        }
                    }
                    else {
                        return j;
                    }
                }
            }
            return j;
        }
        xs(a, b, c) {
            if (this.dn == null) {
                return null;
            }
            let d = this.dn.count;
            let e = this.xn.b0.c;
            let f = this.xn.b0.d;
            let g = this.xn.b0.b;
            let h = this.w6;
            let i = -1;
            let j = -1;
            if (h.cn) {
                let k = this.w9();
                let l = this.y7(b);
                if (k.m != null && l >= 0 && l < k.m.count) {
                    l = truncate(Math.round(k.m._inner[l]));
                }
                i = l;
                j = l;
            }
            else {
                let m = this.j3(b);
                i = truncate(Math.floor((intDivide(m, g)))) * g;
                j = i + (g - 1);
            }
            if (c != null && c.g == i && c.f == j) {
                return c;
            }
            let n = new HighlightingInfo();
            n.i = this;
            n.g = i;
            n.f = j;
            return n;
        }
        xr(a, b, c) {
            let d = new HighlightingInfo();
            d.i = this;
            d.g = 0;
            d.f = this.dn.count - 1;
            d.c = true;
            if (c != null && c.g == d.g && c.f == d.f) {
                return c;
            }
            return d;
        }
        ya() {
            return this.yc;
        }
        ir(a) {
            let b = this.cv.by;
            let c = this.cv.bx;
            let d = -1;
            if (!b.isEmpty && !c.isEmpty && this.w6 != null) {
                let e = this.getEffectiveViewport1(this.cv);
                let f = new ScalerParams(0, b, c, this.w6.ch, e);
                let g = this.wj(this.cv);
                let h = this.w6.eo(g.left, f);
                let i = this.w6.eo(g.right, f);
                let j = (a.x - b.left) / b.width;
                let k = h + (j * (i - h));
                if (this.w6.categoryMode != 0) {
                    k -= 0.5;
                }
                d = k;
            }
            return d;
        }
        j3(a) {
            return truncate(Math.round(this.ir(a)));
        }
        get_fs() {
            return typeCast(ISortingAxis_$type, this.w6) !== null;
        }
        a0() {
            return this.w6;
        }
        j5(a) {
            let b = this.cv.by;
            let c = this.cv.bx;
            let d = this.getEffectiveViewport1(this.cv);
            if (b.isEmpty || c.isEmpty) {
                return -1;
            }
            let e = new Rect(0, 0, 0, 1, 1);
            let f = new ScalerParams(0, b, c, this.w6.ch, d);
            let g = typeCast(ISortingAxis_$type, this.w6);
            let h = this.wj(this.cv);
            let i = h.left;
            let j = h.right;
            let k = this.v7(a).x;
            let l = this.w6.eo(k, f);
            let m = this.w6.eo(i, f);
            let n = this.w6.eo(j, f);
            l = Math.max(m, Math.min(n, l));
            let o = g.i$f(l);
            return o;
        }
        kn(a) {
            let b = 0;
            if (typeCast(ISortingAxis_$type, this.w6) !== null) {
                b = this.j5(a);
                if (b == -1) {
                    return null;
                }
            }
            else {
                b = this.j3(a);
            }
            return b >= 0 && this.dn != null && b < this.dn.count ? this.dn.item(b) : null;
        }
        w9() {
            let a = this.w8;
            if (this.d3()) {
                if (this.dc.g) {
                    this.dc.t();
                }
                a = this.xc;
            }
            return a;
        }
        yx(a, b, c) {
            if (this.w6 == null) {
                return 0;
            }
            return this.w6.k9(a, b, c, this.index);
        }
        gn() {
            return this.yd;
        }
        xg() {
            return 11;
        }
        c4() {
            return 3;
        }
        c7() {
            return 3;
        }
        aae(a, b) {
            this.qx();
            this.wi = a.ab;
            b.aw();
        }
        aac(a, b) {
            a.v();
        }
        he(a, b, c) {
            let d = super.he(a, b, c);
            let e = c;
            if (this.dn == null || this.dn.count == 0 || !c.o() || b.isEmpty || a.isEmpty || this.w6 == null || this.xp == null || e.b0.b < 1 || this.w6.bi == null || this.xp.bi == null || this.xp.l4 == this.xp.l3 || this.w6.itemsSource == null || this.w6.lj < 1) {
                e.b0.b = 0;
                d = false;
            }
            return d;
        }
        getViewInfo(a, b) {
            b = this.cv.by;
            a = this.cv.bx;
            return {
                p0: a,
                p1: b
            };
        }
        y1(a) {
            let b = 0;
            let c = a.by;
            let d = a.bx;
            let e = this.getEffectiveViewport1(this.cv);
            if (!c.isEmpty && !d.isEmpty && this.xp != null) {
                let f = new ScalerParams(0, c, d, this.xp.ch, e);
                f.b = this.iy();
                b = this.xp.el(this.xp.me, f);
            }
            return b;
        }
        o6() {
            super.o6();
            this.xa.w();
            this.w8.w();
            this.xc.w();
        }
        rs(a) {
            super.rs(a);
            this.xn.b0.e(this.id);
            if (this.d5(this.cv)) {
                return;
            }
            if (this.gr) {
                if (this.d3()) {
                    this.aae(this.xc, this.xn);
                }
                else {
                    this.aae(this.w8, this.xn);
                }
                return;
            }
            if (this.gf(a)) {
                let b = this.xa;
                if (this.d3()) {
                    if (this.dc.g) {
                        this.dc.t();
                    }
                    this.xa = this.xc;
                    this.xc = b;
                }
                else {
                    this.xa = this.w8;
                    this.w8 = b;
                }
                this.aac(this.w8, this.xn);
                let c = false;
                if (this.g3) {
                    this.dc.w();
                    this.dc.r = truncate(this.ja());
                    this.dc.b = this.de != null ? this.de : this.dd;
                    if (this.g4) {
                        let d = this.xc;
                        this.xc = this.xa;
                        this.xa = d;
                        c = true;
                    }
                    this.xi.s(this, this.xa, this.w8, this.isVertical, this.w6, this.xp, this.xh, this.xg(), this.c5, this.c4(), () => this.y1(this.xn), this.xn.bx);
                    if (this.g4) {
                        let e = this.xc;
                        this.xc = this.xa;
                        this.xa = e;
                    }
                    this.g4 = false;
                }
                if (this.g6) {
                    this.dc.w();
                    this.dc.r = truncate(this.jb());
                    this.dc.b = this.df != null ? this.df : this.dd;
                    if (this.g1) {
                        let f = this.xc;
                        this.xc = this.xa;
                        this.xa = f;
                        c = true;
                    }
                    this.xi.s(this, this.xa, this.w8, this.isVertical, this.w6, this.xp, this.xh, this.xg(), this.c5, this.c4(), () => this.y1(this.xn), this.xn.bx);
                    let g = this.w8;
                    this.w8 = this.xa;
                    this.xa = g;
                    if (this.g1) {
                        let h = this.xc;
                        this.xc = this.xa;
                        this.xa = h;
                    }
                    this.g1 = false;
                }
                this.d9 = false;
                if (!c) {
                    this.o1();
                }
                this.r7();
                if (this.g3) {
                    this.g2 = false;
                    this.g5 = false;
                    this.gk = false;
                    this.g1 = true;
                }
                if (this.g6) {
                    this.g5 = false;
                    this.g4 = true;
                }
            }
            else {
                this.aac(this.w8, this.xn);
                this.aae(this.w8, this.xn);
            }
        }
        go() {
            return true;
        }
        yf(a) {
            for (let b of fromEnum(a)) {
                switch (b) {
                    case "HighColumn":
                        if (this.x0 == null) {
                            return false;
                        }
                        break;
                    case "LowColumn":
                        if (this.x1 == null) {
                            return false;
                        }
                        break;
                    case "OpenColumn":
                        if (this.x2 == null) {
                            return false;
                        }
                        break;
                    case "CloseColumn":
                        if (this.xz == null) {
                            return false;
                        }
                        break;
                    case "VolumeColumn":
                        if (this.x4 == null) {
                            return false;
                        }
                        break;
                }
            }
            return true;
        }
        get yg() {
            return this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null;
        }
        yb() {
            return this.typical != null;
        }
        z4(a) {
            this.typical(this, a);
        }
        get yo() {
            return this.yl();
        }
        *_yl() {
            let a = false;
            if (this.yb()) {
                let b = this.xj(0, this.dn.count);
                this.z4(new FinancialEventArgs(0, this.dn.count, b, this.xk(b), this));
                if (b.typicalColumn != null) {
                    a = true;
                    for (let c of fromEnum(b.typicalColumn)) {
                        yield c;
                    }
                }
            }
            if (!a) {
                if (this.yg && this.w6.i$b != null) {
                    let d = this.dn.count;
                    let e = this.w6.i$b;
                    for (let f = 0; f < d; ++f) {
                        yield (this.x0.item(e.item(f)) + this.x1.item(e.item(f)) + this.xz.item(e.item(f))) / 3;
                    }
                }
                else {
                    let g = this.dn.count;
                    for (let h = 0; h < g; ++h) {
                        yield (this.x0.item(h) + this.x1.item(h) + this.xz.item(h)) / 3;
                    }
                }
            }
        }
        yl() {
            return toEnum(() => this._yl());
        }
        get yn() {
            return this.yk();
        }
        *_yk() {
            if (this.yg && this.w6.i$b != null) {
                let a = this.x0.count;
                let b = this.w6.i$b;
                if (a > 0) {
                    yield this.y2(this.x0.item(b.item(0)) - this.x1.item(b.item(0)));
                }
                for (let c = 1; c < a; ++c) {
                    yield Math.max(this.y2(this.x0.item(b.item(c)) - this.x1.item(b.item(c))), Math.max(this.y2(Math.abs(this.x0.item(b.item(c)) - this.xz.item(b.item(c - 1)))), this.y2(Math.abs(this.x1.item(b.item(c)) - this.xz.item(b.item(c - 1))))));
                }
            }
            else {
                let d = this.x0.count;
                if (d > 0) {
                    yield this.y2(this.x0.item(0) - this.x1.item(0));
                }
                for (let e = 1; e < d; ++e) {
                    yield Math.max(this.y2(this.x0.item(e) - this.x1.item(e)), Math.max(this.y2(Math.abs(this.x0.item(e) - this.xz.item(e - 1))), this.y2(Math.abs(this.x1.item(e) - this.xz.item(e - 1)))));
                }
            }
        }
        yk() {
            return toEnum(() => this._yk());
        }
        get ym() {
            return this.yj();
        }
        *_yj() {
            if (this.yg && this.w6.i$b != null) {
                let a = this.x1.count;
                let b = this.w6.i$b;
                if (a > 0) {
                    yield this.y2(this.x1.item(b.item(0)));
                }
                for (let c = 1; c < a; c++) {
                    yield Math.min(this.y2(this.x1.item(b.item(c))), this.y2(this.xz.item(b.item(c - 1))));
                }
            }
            else {
                let d = this.x1.count;
                if (d > 0) {
                    yield this.y2(this.x1.item(0));
                }
                for (let e = 1; e < d; e++) {
                    yield Math.min(this.y2(this.x1.item(e)), this.y2(this.xz.item(e - 1)));
                }
            }
        }
        yj() {
            return toEnum(() => this._yj());
        }
        yq(a) {
            if (a == null) {
                return null;
            }
            if (this.yg && this.w6.i$b != null) {
                return new SafeSortedReadOnlyDoubleCollection(0, a, this.w6.i$b);
            }
            return new SafeReadOnlyDoubleCollection(0, a);
        }
        yp() {
            let a = new List$1(String_$type, 0);
            a.add("HighColumn");
            a.add("LowColumn");
            a.add("CloseColumn");
            if (this.typicalBasedOn != null && this.dn != null && !this.x5) {
                this.x5 = true;
                let b = this.xj(0, this.dn.count);
                this.x5 = false;
                let c = new FinancialEventArgs(0, this.dn.count, b, this.xk(b), this);
                this.typicalBasedOn(this, c);
                if (c.basedOn != null && c.basedOn.count > 0) {
                    return c.basedOn;
                }
            }
            return a;
        }
        getOpenColumnAsArray() {
            if (this.x2 == null) {
                return null;
            }
            if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
                if (this.yu != null) {
                    return this.yu.asArray();
                }
                let a = new List$1(Number_$type, 2, this.x2.count);
                let b = this.w6.i$b;
                let c = this.x2.asArray();
                for (let d = 0; d < b.count; d++) {
                    a.add(c[b._inner[d]]);
                }
                this.yu = a;
                return this.yu.asArray();
            }
            else {
                return this.x2.asArray();
            }
        }
        getHighColumnAsArray() {
            if (this.x0 == null) {
                return null;
            }
            if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
                if (this.ys != null) {
                    return this.ys.asArray();
                }
                let a = new List$1(Number_$type, 2, this.x0.count);
                let b = this.w6.i$b;
                let c = this.x0.asArray();
                for (let d = 0; d < b.count; d++) {
                    a.add(c[b._inner[d]]);
                }
                this.ys = a;
                return this.ys.asArray();
            }
            else {
                return this.x0.asArray();
            }
        }
        getLowColumnAsArray() {
            if (this.x1 == null) {
                return null;
            }
            if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
                if (this.yt != null) {
                    return this.yt.asArray();
                }
                let a = new List$1(Number_$type, 2, this.x1.count);
                let b = this.w6.i$b;
                let c = this.x1.asArray();
                for (let d = 0; d < b.count; d++) {
                    a.add(c[b._inner[d]]);
                }
                this.yt = a;
                return this.yt.asArray();
            }
            else {
                return this.x1.asArray();
            }
        }
        getCloseColumnAsArray() {
            if (this.xz == null) {
                return null;
            }
            if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
                if (this.yr != null) {
                    return this.yr.asArray();
                }
                let a = new List$1(Number_$type, 2, this.xz.count);
                let b = this.w6.i$b;
                let c = this.xz.asArray();
                for (let d = 0; d < b.count; d++) {
                    a.add(c[b._inner[d]]);
                }
                this.yr = a;
                return this.yr.asArray();
            }
            else {
                return this.xz.asArray();
            }
        }
        getVolumeColumnAsArray() {
            if (this.x4 == null) {
                return null;
            }
            if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
                if (this.yv != null) {
                    return this.yv.asArray();
                }
                let a = new List$1(Number_$type, 2, this.x4.count);
                let b = this.w6.i$b;
                let c = this.x4.asArray();
                for (let d = 0; d < b.count; d++) {
                    a.add(c[b._inner[d]]);
                }
                this.yv = a;
                return this.yv.asArray();
            }
            else {
                return this.x4.asArray();
            }
        }
        pb(a, b, c, d) {
            super.pb(a, b, c, d);
            if (this.w6 == null || !(typeCast(ISortingAxis_$type, this.w6) !== null)) {
                return;
            }
            let e = this.w6.i$b;
            switch (a) {
                case 1:
                    for (let f = b; f < c; f++) {
                        let g = e._inner[f];
                        if (this.yu != null) {
                            this.yu.insert(f, this.x2.item(g));
                        }
                        if (this.ys != null) {
                            this.ys.insert(f, this.x0.item(g));
                        }
                        if (this.yt != null) {
                            this.yt.insert(f, this.x1.item(g));
                        }
                        if (this.yr != null) {
                            this.yr.insert(f, this.xz.item(g));
                        }
                        if (this.yv != null) {
                            this.yv.insert(f, this.x4.item(g));
                        }
                    }
                    break;
                case 0:
                    if (this.yu != null) {
                        this.yu.v(b, c);
                    }
                    if (this.ys != null) {
                        this.ys.v(b, c);
                    }
                    if (this.yt != null) {
                        this.yt.v(b, c);
                    }
                    if (this.yr != null) {
                        this.yr.v(b, c);
                    }
                    if (this.yv != null) {
                        this.yv.v(b, c);
                    }
                    break;
                case 4:
                    if (this.yu != null) {
                        this.yu = new List$1(Number_$type, 0);
                    }
                    if (this.ys != null) {
                        this.ys = new List$1(Number_$type, 0);
                    }
                    if (this.yt != null) {
                        this.yt = new List$1(Number_$type, 0);
                    }
                    if (this.yr != null) {
                        this.yr = new List$1(Number_$type, 0);
                    }
                    if (this.yv != null) {
                        this.yv = new List$1(Number_$type, 0);
                    }
                    break;
                case 3:
                case 2:
                    for (let h = b; h < c; h++) {
                        let i = e._inner[h];
                        if (this.yu != null) {
                            this.yu._inner[h] = this.x2.item(i);
                        }
                        if (this.ys != null) {
                            this.ys._inner[h] = this.x0.item(i);
                        }
                        if (this.yt != null) {
                            this.yt._inner[h] = this.x1.item(i);
                        }
                        if (this.yr != null) {
                            this.yr._inner[h] = this.xz.item(i);
                        }
                        if (this.yv != null) {
                            this.yv._inner[h] = this.x4.item(i);
                        }
                    }
                    break;
            }
            this.aaf();
            this.sn();
        }
        xj(a, b) {
            let c = this.yq(this.x2);
            let d = this.yq(this.xz);
            let e = this.yq(this.x0);
            let f = this.yq(this.x1);
            let g = this.yq(this.x4);
            let h = ((() => {
                let $ret = new FinancialCalculationDataSource();
                $ret.typicalColumn = new CalculatedColumn(1, new SafeEnumerable(this.yo), this.yp());
                $ret.trueRange = new CalculatedColumn(1, new SafeEnumerable(this.yn), ((() => {
                    let $ret = new List$1(String_$type, 0);
                    $ret.add("HighColumn");
                    $ret.add("LowColumn");
                    $ret.add("CloseColumn");
                    return $ret;
                })()));
                $ret.trueLow = new CalculatedColumn(1, new SafeEnumerable(this.ym), ((() => {
                    let $ret = new List$1(String_$type, 0);
                    $ret.add("LowColumn");
                    $ret.add("CloseColumn");
                    return $ret;
                })()));
                $ret.openColumn = c;
                $ret.closeColumn = d;
                $ret.highColumn = e;
                $ret.lowColumn = f;
                $ret.volumeColumn = g;
                $ret.calculateFrom = a;
                $ret.calculateCount = b;
                $ret.minimumValue = NaN;
                $ret.maximumValue = NaN;
                $ret.count = this.dn != null ? this.dn.count : 0;
                return $ret;
            })());
            return h;
        }
        y2(a) {
            if (isInfinity(a) || isNaN_(a)) {
                return 0;
            }
            return a;
        }
        xk(a) {
            return ((() => {
                let $ret = new FinancialCalculationSupportingCalculations();
                $ret.a = new ColumnSupportingCalculation(1, TrendCalculators.b, new List$1(String_$type, 0));
                $ret.c = new ColumnSupportingCalculation(1, TrendCalculators.e, new List$1(String_$type, 0));
                $ret.d = new ColumnSupportingCalculation(1, TrendCalculators.f, new List$1(String_$type, 0));
                $ret.b = new ColumnSupportingCalculation(1, TrendCalculators.d, new List$1(String_$type, 0));
                $ret.h = new DataSourceSupportingCalculation(1, (b) => TrendCalculators.b(b.volumeColumn, b.shortPeriod), ((() => {
                    let $ret = new List$1(String_$type, 0);
                    $ret.add("VolumeColumn");
                    return $ret;
                })()));
                $ret.f = new DataSourceSupportingCalculation(1, (b) => TrendCalculators.b(b.volumeColumn, b.longPeriod), ((() => {
                    let $ret = new List$1(String_$type, 0);
                    $ret.add("VolumeColumn");
                    return $ret;
                })()));
                $ret.g = new DataSourceSupportingCalculation(1, (b) => TrendCalculators.b(b.typicalColumn, b.shortPeriod), a.typicalColumn.basedOn);
                $ret.e = new DataSourceSupportingCalculation(1, (b) => TrendCalculators.b(b.typicalColumn, b.longPeriod), a.typicalColumn.basedOn);
                $ret.k = Series.hq;
                $ret.j = Series.hp;
                $ret.i = runOn(this, this.y2);
                return $ret;
            })());
        }
        get categoryAxis() {
            return this.w6;
        }
        rt(a, b) {
            super.rt(a, b);
            if (!this.g0) {
                this.cv.a4(b);
                return;
            }
            let c = typeCast(FinancialSeriesView.$, this.cu);
            this.cv.a4(b);
            c.b0.e(this.id);
            if (this.d5(this.cu)) {
                return;
            }
            if (!this.gs) {
                this.aac(this.xb, c);
            }
            this.gs = false;
            this.aae(this.xb, c);
            this.g0 = false;
        }
        rq(a, b, c, d, e) {
            super.rq(a, b, c, d, e);
            let f = this.ba.a$j.item(d);
            let g = f;
            g.b0.e(this.id);
            f.a3(c);
            if (this.d5(g)) {
                return;
            }
            if (this.w7 == null) {
                this.w7 = new CategoryFrame(3);
            }
            this.w7.v();
            this.aac(this.w7, g);
            this.aae(this.w7, g);
        }
        aw(a) {
            if (this.w6 == a) {
                return this.xp;
            }
            if (this.xp == a) {
                return this.w6;
            }
            return null;
        }
        preferredCategoryMode(a) {
            let b = typeCast(IDateTimeAxis_$type, this.w6);
            return b != null ? b.preferredCategoryMode(this) : this.xe;
        }
        get xe() {
            return this.xd;
        }
        set xe(a) {
            let b = this.xe;
            if (a != b) {
                this.xd = a;
                this.raisePropertyChanged("CategoryMode", enumGetBox(CategoryMode_$type, b), enumGetBox(CategoryMode_$type, this.xe));
            }
        }
        get_fm() {
            return true;
        }
        r3(a, b) {
            super.r3(a, b);
            this.aaw = a;
        }
        aaf() {
            this.im = this.xz == null || this.xz.count == 0 ? NaN : this.xz.item(this.xz.count - 1);
        }
        kp(a) {
            if (this.xo != null && a != null) {
                if (this.xq == null || this.xq.propertyName != this.xo.dateTimeMemberPath) {
                    this.xq = new FastReflectionHelper(false, this.xo.dateTimeMemberPath);
                }
                let b = this.xq.getPropertyValue(a);
                if (typeCast(Date_$type, b) !== null) {
                    return dateToStringFormat(b, this.xo.itemLabelFormat());
                }
            }
            return super.kp(a);
        }
    }
    FinancialSeries.$t = /*@__PURE__*/ markType(FinancialSeries, 'FinancialSeries', Series.$, [IHasCategoryAxis_$type, IHasCategoryModePreference_$type]);
    FinancialSeries.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, FinancialSeries, 'raisePropertyChanged', ['CloseMemberPath:za:aag', [2, null], 'HighMemberPath:zn:aam', [2, null], 'HighlightedCloseMemberPath:zd:aah', [2, null], 'HighlightedHighMemberPath:zf:aai', [2, null], 'HighlightedLowMemberPath:zh:aaj', [2, null], 'HighlightedOpenMemberPath:zj:aak', [2, null], 'HighlightedVolumeMemberPath:zl:aal', [2, null], 'IsCustomCategoryStyleAllowed:yc:aan', [0, false], 'IsTransitionInEnabled:yd:aao', [0, false], 'LowMemberPath:zs:aap', [2, null], 'NegativeBrush:aaw:aaq', [Brush.$, null], 'OpenMemberPath:zw:aar', [2, null], 'TransitionInMode:xh:aas', [CategoryTransitionInMode_$type, /*@__PURE__*/ enumGetBox(CategoryTransitionInMode_$type, 0)], 'VolumeMemberPath:z0:aat', [2, null], 'XAxis:w6:aau', [CategoryAxisBase.$, null], 'YAxis:xp:aav', [NumericYAxis.$, null]]);
    return FinancialSeries;
})();
/**
 * @hidden
 */
export let FinancialCalculationSupportingCalculations = /*@__PURE__*/ (() => {
    class FinancialCalculationSupportingCalculations extends Base {
        constructor() {
            super(...arguments);
            this.a = null;
            this.c = null;
            this.d = null;
            this.b = null;
            this.h = null;
            this.f = null;
            this.g = null;
            this.e = null;
            this.k = null;
            this.j = null;
            this.i = null;
        }
    }
    FinancialCalculationSupportingCalculations.$t = /*@__PURE__*/ markType(FinancialCalculationSupportingCalculations, 'FinancialCalculationSupportingCalculations');
    return FinancialCalculationSupportingCalculations;
})();
/**
 * @hidden
 */
export let DataSourceSupportingCalculation = /*@__PURE__*/ (() => {
    class DataSourceSupportingCalculation extends SupportingCalculation$1 {
        constructor(a, ..._rest) {
            a = (a == void 0) ? 0 : a;
            switch (a) {
                case 0:
                    {
                        let c = _rest[0];
                        super(Delegate_$type, 0, c);
                    }
                    break;
                case 1:
                    {
                        let c = _rest[0];
                        let d = _rest[1];
                        super(Delegate_$type, 1, c, d);
                    }
                    break;
            }
        }
    }
    DataSourceSupportingCalculation.$t = /*@__PURE__*/ markType(DataSourceSupportingCalculation, 'DataSourceSupportingCalculation', /*@__PURE__*/ SupportingCalculation$1.$.specialize(Delegate_$type));
    return DataSourceSupportingCalculation;
})();
/**
 * @hidden
 */
export let FinancialCalculationDataSource = /*@__PURE__*/ (() => {
    class FinancialCalculationDataSource extends Base {
        constructor() {
            super(...arguments);
            this._openColumn = null;
            this._closeColumn = null;
            this._highColumn = null;
            this._lowColumn = null;
            this._volumeColumn = null;
            this._indicatorColumn = null;
            this._typicalColumn = null;
            this._trueRange = null;
            this._trueLow = null;
            this._period = 0;
            this._shortPeriod = 0;
            this._longPeriod = 0;
            this._count = 0;
            this._calculateFrom = 0;
            this._calculateCount = 0;
            this._multiplier = 0;
            this._minimumValue = 0;
            this._maximumValue = 0;
            this._specifiesRange = false;
            this._series = null;
        }
        get openColumn() {
            return this._openColumn;
        }
        set openColumn(a) {
            this._openColumn = a;
        }
        get closeColumn() {
            return this._closeColumn;
        }
        set closeColumn(a) {
            this._closeColumn = a;
        }
        get highColumn() {
            return this._highColumn;
        }
        set highColumn(a) {
            this._highColumn = a;
        }
        get lowColumn() {
            return this._lowColumn;
        }
        set lowColumn(a) {
            this._lowColumn = a;
        }
        get volumeColumn() {
            return this._volumeColumn;
        }
        set volumeColumn(a) {
            this._volumeColumn = a;
        }
        get indicatorColumn() {
            return this._indicatorColumn;
        }
        set indicatorColumn(a) {
            this._indicatorColumn = a;
        }
        get typicalColumn() {
            return this._typicalColumn;
        }
        set typicalColumn(a) {
            this._typicalColumn = a;
        }
        get trueRange() {
            return this._trueRange;
        }
        set trueRange(a) {
            this._trueRange = a;
        }
        get trueLow() {
            return this._trueLow;
        }
        set trueLow(a) {
            this._trueLow = a;
        }
        get period() {
            return this._period;
        }
        set period(a) {
            this._period = a;
        }
        get shortPeriod() {
            return this._shortPeriod;
        }
        set shortPeriod(a) {
            this._shortPeriod = a;
        }
        get longPeriod() {
            return this._longPeriod;
        }
        set longPeriod(a) {
            this._longPeriod = a;
        }
        get count() {
            return this._count;
        }
        set count(a) {
            this._count = a;
        }
        get calculateFrom() {
            return this._calculateFrom;
        }
        set calculateFrom(a) {
            this._calculateFrom = a;
        }
        get calculateCount() {
            return this._calculateCount;
        }
        set calculateCount(a) {
            this._calculateCount = a;
        }
        get multiplier() {
            return this._multiplier;
        }
        set multiplier(a) {
            this._multiplier = a;
        }
        get minimumValue() {
            return this._minimumValue;
        }
        set minimumValue(a) {
            this._minimumValue = a;
        }
        get maximumValue() {
            return this._maximumValue;
        }
        set maximumValue(a) {
            this._maximumValue = a;
        }
        get specifiesRange() {
            return this._specifiesRange;
        }
        set specifiesRange(a) {
            this._specifiesRange = a;
        }
        get series() {
            return this._series;
        }
        set series(a) {
            this._series = a;
        }
    }
    FinancialCalculationDataSource.$t = /*@__PURE__*/ markType(FinancialCalculationDataSource, 'FinancialCalculationDataSource');
    return FinancialCalculationDataSource;
})();
/**
 * @hidden
 */
export let FinancialSeriesView = /*@__PURE__*/ (() => {
    class FinancialSeriesView extends SeriesView {
        b3(a) {
            if (this.b4 != a.u) {
                return true;
            }
            return false;
        }
        b5(a) {
            this.b4 = a.u;
        }
        constructor(a) {
            super(a);
            this.b2 = null;
            this.b0 = null;
            this.b4 = 0;
            this.b4 = -1;
            this.b2 = a;
            this.b0 = this.b1();
        }
        au() {
            super.au();
            if (!this.q) {
            }
        }
        getDefaultTooltipTemplate() {
            let a = "<div class='ui-chart-default-tooltip-content'>";
            if (this.b2.w6.ce) {
                a += "<span>${itemLabel}</span><br/>";
            }
            else if (this.b2.w6.f2 != null) {
                a += "<span>${item." + this.b2.w6.f2 + "}</span><br/>";
            }
            a += "<span class='ui-priority-primary'";
            let b = this.e.ba.a$u(this.e);
            if (!stringIsNullOrEmpty(b)) {
                a += " style='color:" + b + "'";
            }
            let c = new SRProvider(this.w);
            let d = c.b1("financialSeries_Open");
            let e = c.b1("financialSeries_High");
            let f = c.b1("financialSeries_Low");
            let g = c.b1("financialSeries_Close");
            d = stringIsNullOrEmpty(d) ? "Open" : d;
            e = stringIsNullOrEmpty(e) ? "High" : e;
            f = stringIsNullOrEmpty(f) ? "Low" : f;
            g = stringIsNullOrEmpty(g) ? "Close" : g;
            a += ">" + this.b2.title + "</span><table><tr><td>" + d + ":</td><td>" + "${item." + this.b2.zw + "}</td></tr><tr><td>" + e + ":</td><td>" + "${item." + this.b2.zn + "}</td></tr><tr><td>" + f + ":</td><td>" + "${item." + this.b2.zs + "}</td></tr><tr><td>" + g + ":</td><td>" + "${item." + this.b2.za + "}</td></tr>";
            if (!stringIsNullOrEmpty(this.b2.z0)) {
                let h = c.b1("financialSeries_Volume");
                h = stringIsNullOrEmpty(h) ? "Volume" : h;
                a += "<tr><td>" + h + ":</td><td>" + "${item." + this.b2.z0 + "}</td></tr>";
            }
            a += "</table></div>";
            return a;
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.v1 = a;
            this.e.jm = 5;
            this.e.jn = 2;
            this.e.jo = 2;
        }
    }
    FinancialSeriesView.$t = /*@__PURE__*/ markType(FinancialSeriesView, 'FinancialSeriesView', SeriesView.$);
    return FinancialSeriesView;
})();
