// igniteui-react-charts/esm2015/lib/FinancialPriceSeriesProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { IFinancialPriceSeriesProxy_$type } from "./IFinancialPriceSeriesProxy";
/**
 * @hidden
 */
export let FinancialPriceSeriesProxy = /*@__PURE__*/ (() => {
    class FinancialPriceSeriesProxy extends Base {
        getCloseMemberPath(a) {
            return a.za;
        }
        getHighMemberPath(a) {
            return a.zn;
        }
        getLowMemberPath(a) {
            return a.zs;
        }
        getOpenMemberPath(a) {
            return a.zw;
        }
        getVolumeMemberPath(a) {
            return a.z0;
        }
        matchesType(a) {
            if (!a.e6) {
                return false;
            }
            if (a.e7 || a.e8) {
                return false;
            }
            return true;
        }
        setCategoryMode(a, b) {
            a.xe = b;
        }
        setCloseMemberPath(a, b) {
            a.za = b;
        }
        setDisplayType(a, b) {
            a.aa3 = b;
        }
        setHighMemberPath(a, b) {
            a.zn = b;
        }
        setLowMemberPath(a, b) {
            a.zs = b;
        }
        setOpenMemberPath(a, b) {
            a.zw = b;
        }
        setVolumeMemberPath(a, b) {
            a.z0 = b;
        }
        setXAxis(a, b) {
            a.w6 = b;
        }
        setYAxis(a, b) {
            a.xp = b;
        }
        setTrendLineType(a, b) {
            if (a.e9) {
                a.trendLineType = b;
            }
        }
        getTrendLineType(a) {
            if (a.e9) {
                return a.trendLineType;
            }
            return 0;
        }
    }
    FinancialPriceSeriesProxy.$t = /*@__PURE__*/ markType(FinancialPriceSeriesProxy, 'FinancialPriceSeriesProxy', Base.$, [IFinancialPriceSeriesProxy_$type]);
    return FinancialPriceSeriesProxy;
})();
