// igniteui-react-charts/esm2015/lib/FinancialChartDropDownControl.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { TemplatedHtmlControl } from "./TemplatedHtmlControl";
import { EventArgs, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let FinancialChartDropDownControl = /*@__PURE__*/ (() => {
    class FinancialChartDropDownControl extends TemplatedHtmlControl {
        constructor(a, b, c) {
            super(a, b, c);
            this.selectionChanged = null;
            this.o = null;
            this.p = null;
        }
        v() {
            if (this.selectionChanged != null) {
                this.selectionChanged(this, EventArgs.empty);
            }
        }
        l(a) {
            super.l(a);
            this.o.f(this.d, this.p, this.s, this.t);
        }
        get_q() {
            return false;
        }
        get q() {
            return this.get_q();
        }
        u(a) {
            if (this.p.isOpen) {
                let b = a.target;
                let e_ = a;
                if ((e_.composed)) {
                    let c = (e_.composedPath());
                    if (c != null && c.length > 0) {
                        b = c[0];
                    }
                }
                let d = b.matches(this.s + " *");
                let e = d || (this.q && b.matches(this.t + " *"));
                if (!e) {
                    this.p.isOpen = false;
                    this.p.updateUI();
                }
            }
        }
    }
    FinancialChartDropDownControl.$t = /*@__PURE__*/ markType(FinancialChartDropDownControl, 'FinancialChartDropDownControl', TemplatedHtmlControl.$);
    return FinancialChartDropDownControl;
})();
