// igniteui-react-charts/esm2015/lib/FinancialChartDropDownControlView.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { SRProvider } from "igniteui-react-core";
/**
 * @hidden
 */
export let FinancialChartDropDownControlView = /*@__PURE__*/ (() => {
    class FinancialChartDropDownControlView extends Base {
        constructor() {
            super(...arguments);
            this.b = false;
            this.c = false;
            this.d = null;
            this.e = null;
            this.a = null;
        }
        i(a) {
            if (!this.b && !this.c) {
                this.g();
            }
            this.b = false;
            this.c = false;
        }
        h(a) {
            this.b = true;
        }
        k(a) {
        }
        j(a) {
            this.c = true;
            this.k(a.target);
        }
        f(a, b, c, d) {
            this.a = b;
            this.a.a = new SRProvider(a);
        }
        g() {
            this.a.isOpen = false;
            this.a.updateUI();
        }
    }
    FinancialChartDropDownControlView.$t = /*@__PURE__*/ markType(FinancialChartDropDownControlView, 'FinancialChartDropDownControlView');
    return FinancialChartDropDownControlView;
})();
