// igniteui-react-charts/esm2015/lib/FinancialChartDropDownControlViewModel.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let FinancialChartDropDownControlViewModel = /*@__PURE__*/ (() => {
    class FinancialChartDropDownControlViewModel extends Base {
        constructor() {
            super(...arguments);
            this._onSelectionChanged = null;
            this._isOpen = false;
            this._notifyChanged = null;
            this.a = null;
            this.b = null;
        }
        get onSelectionChanged() {
            return this._onSelectionChanged;
        }
        set onSelectionChanged(a) {
            this._onSelectionChanged = a;
        }
        get isOpen() {
            return this._isOpen;
        }
        set isOpen(a) {
            this._isOpen = a;
        }
        get notifyChanged() {
            return this._notifyChanged;
        }
        set notifyChanged(a) {
            this._notifyChanged = a;
        }
        updateUI() {
            if (this.notifyChanged != null) {
                this.notifyChanged();
            }
        }
        onButtonPressed() {
            this.isOpen = !this.isOpen;
            this.updateUI();
        }
        handleDropDownClick(a) {
            if (this.b != null) {
                this.b(a);
            }
        }
    }
    FinancialChartDropDownControlViewModel.$t = /*@__PURE__*/ markType(FinancialChartDropDownControlViewModel, 'FinancialChartDropDownControlViewModel');
    return FinancialChartDropDownControlViewModel;
})();
