// igniteui-react-charts/esm2015/lib/FinancialChartContext.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let FinancialChartContext = /*@__PURE__*/ (() => {
    class FinancialChartContext extends Base {
        constructor() {
            super(...arguments);
            this._vm = null;
            this._template = null;
            this._container = null;
            this._toolbarContainer = null;
            this._legendContainer = null;
            this._indicatorsContainer = null;
            this._volumeContainer = null;
            this._zoomContainer = null;
        }
        get vm() {
            return this._vm;
        }
        set vm(a) {
            this._vm = a;
        }
        get template() {
            return this._template;
        }
        set template(a) {
            this._template = a;
        }
        get container() {
            return this._container;
        }
        set container(a) {
            this._container = a;
        }
        get toolbarContainer() {
            return this._toolbarContainer;
        }
        set toolbarContainer(a) {
            this._toolbarContainer = a;
        }
        get legendContainer() {
            return this._legendContainer;
        }
        set legendContainer(a) {
            this._legendContainer = a;
        }
        get indicatorsContainer() {
            return this._indicatorsContainer;
        }
        set indicatorsContainer(a) {
            this._indicatorsContainer = a;
        }
        get volumeContainer() {
            return this._volumeContainer;
        }
        set volumeContainer(a) {
            this._volumeContainer = a;
        }
        get zoomContainer() {
            return this._zoomContainer;
        }
        set zoomContainer(a) {
            this._zoomContainer = a;
        }
    }
    FinancialChartContext.$t = /*@__PURE__*/ markType(FinancialChartContext, 'FinancialChartContext');
    return FinancialChartContext;
})();
