// igniteui-react-charts/esm2015/lib/ForceIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ForceIndexIndicatorStrategy } from "./ForceIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ForceIndexIndicator = /*@__PURE__*/ (() => {
    class ForceIndexIndicator extends StrategyBasedIndicator {
        bj() {
            return new ForceIndexIndicator();
        }
        get_aa2() {
            return 1;
        }
        constructor() {
            super();
        }
        get_abv() {
            return new ForceIndexIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return ForceIndexIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(ForceIndexIndicator.ach));
        }
        set acg(a) {
            this.h(ForceIndexIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    ForceIndexIndicator.$t = /*@__PURE__*/ markType(ForceIndexIndicator, 'ForceIndexIndicator', StrategyBasedIndicator.$);
    ForceIndexIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(0, ForceIndexIndicator.$);
    return ForceIndexIndicator;
})();
