// igniteui-react-charts/esm2015/lib/FunnelChartSelectedItemsCollection.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ObservableCollection$1 } from "igniteui-react-core";
import { Base, PropertyChangedEventArgs, delegateCombine, markType } from "igniteui-react-core";
import { NotifyCollectionChangedEventArgs } from "igniteui-react-core";
/**
 * @hidden
 */
export let FunnelChartSelectedItemsCollection = /*@__PURE__*/ (() => {
    class FunnelChartSelectedItemsCollection extends ObservableCollection$1 {
        constructor() {
            super(Base.$, 0);
            this.af = false;
        }
        get ae() {
            return this;
        }
        ag(a) {
            this.collectionChanged = delegateCombine(this.collectionChanged, a);
        }
        u(a) {
            let b = this._inner[a];
            if (a == 0) {
                this._inner.shift();
            }
            else {
                this._inner.splice(a, 1);
            }
            this.ad(new PropertyChangedEventArgs("Count"));
            this.ad(new PropertyChangedEventArgs("Item[]"));
            if (!this.af) {
                this.ac(new NotifyCollectionChangedEventArgs(1, 1, b, a));
            }
        }
    }
    FunnelChartSelectedItemsCollection.$t = /*@__PURE__*/ markType(FunnelChartSelectedItemsCollection, 'FunnelChartSelectedItemsCollection', /*@__PURE__*/ ObservableCollection$1.$.specialize(Base.$));
    return FunnelChartSelectedItemsCollection;
})();
