// igniteui-react-charts/esm2015/lib/FunnelSliceEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { EventArgs, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let FunnelSliceEventArgs = /*@__PURE__*/ (() => {
    class FunnelSliceEventArgs extends EventArgs {
        constructor() {
            super(...arguments);
            this.b = 0;
            this.h = null;
            this.d = null;
            this.f = null;
        }
        get index() {
            return this.b;
        }
        set index(a) {
            this.b = a;
        }
        get bounds() {
            return this.h;
        }
        set bounds(a) {
            this.h = a;
        }
        get item() {
            return this.d;
        }
        set item(a) {
            this.d = a;
        }
        get position() {
            return this.f;
        }
        set position(a) {
            this.f = a;
        }
    }
    FunnelSliceEventArgs.$t = /*@__PURE__*/ markType(FunnelSliceEventArgs, 'FunnelSliceEventArgs', EventArgs.$);
    return FunnelSliceEventArgs;
})();
