// igniteui-react-charts/esm2015/lib/HorizontalRangeCategorySeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RangeCategorySeries } from "./RangeCategorySeries";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { DependencyProperty } from "igniteui-react-core";
import { NumericYAxis } from "./NumericYAxis";
import { Axis } from "./Axis";
import { Point_$type, runOn, typeCast, markType, markDep } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { RangeValueList } from "./RangeValueList";
import { ScalerParams } from "./ScalerParams";
import { CategoryXAxis } from "./CategoryXAxis";
import { DataLegendSeriesValueInfo } from "igniteui-react-core";
import { DataLegendSeriesContext } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
/**
 * @hidden
 */
export let HorizontalRangeCategorySeries = /*@__PURE__*/ (() => {
    class HorizontalRangeCategorySeries extends RangeCategorySeries {
        fetchXAxis() {
            return this.xAxis;
        }
        fetchYAxis() {
            return this.yAxis;
        }
        o2() {
            super.o2();
            this.xAxis = null;
            this.yAxis = null;
        }
        aj(a) {
            let b = this.an(this.ys(), this.yu.f, this.zv(this.cv), this.zw(this.cv), this.wh(a), true);
            if (b == null) {
                return null;
            }
            let c = b[0];
            let d = b[1];
            let e = new Array(1);
            let f = new Array(2);
            f[0] = { $type: Point_$type, x: c[0], y: c[2] };
            f[1] = { $type: Point_$type, x: d[0], y: d[2] };
            e[0] = f;
            return e;
        }
        al(a) {
            let b = this.an(this.ys(), this.yu.f, this.zv(this.cv), this.zw(this.cv), this.wh(a), true);
            if (b == null) {
                return null;
            }
            let c = b[0];
            let d = b[1];
            let e = new Array(1);
            let f = new Array(2);
            f[0] = { $type: Point_$type, x: c[0], y: c[1] };
            f[1] = { $type: Point_$type, x: d[0], y: d[1] };
            e[0] = f;
            return e;
        }
        get_fu() {
            return true;
        }
        b2() {
            if (this.b3 == 0) {
                return 1;
            }
            else {
                return super.b2();
            }
        }
        getOffsetValue() {
            return this.RangeFramePreparer.r(this.xAxis, this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
        }
        getCategoryWidth() {
            return this.xAxis.getCategorySize(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
        }
        j8(a, b) {
            return this.j9(a, b, this.xAxis, runOn(this, this.zt), new RangeValueList(this.highColumn, this.lowColumn));
        }
        ka(a, b) {
            return this.kb(a, b, this.xAxis, runOn(this, this.zt), new RangeValueList(this.highColumn, this.lowColumn));
        }
        ip(a, b, c, d, e) {
            if (c == null) {
                return Number.POSITIVE_INFINITY;
            }
            let f = this.xAxis.lj;
            return this.iq(a, b, this.xAxis, d, e, f, runOn(this, this.zt));
        }
        i5(a, b, c) {
            if (this.db == null) {
                return NaN;
            }
            let d = this.getEffectiveViewport1(this.cv);
            let e = new ScalerParams(0, this.db.v4, this.cv.bx, this.xAxis.ch, d);
            let f = this.RangeFramePreparer.r(this.xAxis, this.db.v4, this.cv.bx, d);
            return this.i7(new RangeValueList(this.highColumn, this.lowColumn), a, this.xAxis, e, f, runOn(this, this.zt), b, c);
        }
        i3(a, b, c) {
            if (this.db == null) {
                return NaN;
            }
            let d = this.getEffectiveViewport1(this.cv);
            let e = new ScalerParams(0, this.db.v4, this.cv.bx, this.xAxis.ch, d);
            let f = this.RangeFramePreparer.r(this.xAxis, this.db.v4, this.cv.bx, d);
            return this.i7(this.lowColumn, a, this.xAxis, e, f, runOn(this, this.zt), b, c);
        }
        i1(a, b, c) {
            if (this.db == null) {
                return NaN;
            }
            let d = this.getEffectiveViewport1(this.cv);
            let e = new ScalerParams(0, this.db.v4, this.cv.bx, this.xAxis.ch, d);
            let f = this.RangeFramePreparer.r(this.xAxis, this.db.v4, this.cv.bx, d);
            return this.i7(this.highColumn, a, this.xAxis, e, f, runOn(this, this.zt), b, c);
        }
        v8(a, b, c) {
            return this.we(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.yAxis, this.xAxis, runOn(this, this.i1), (d, e) => this.kb(d, e, this.xAxis, runOn(this, this.zt), this.highColumn), (d, e) => this.j9(d, e, this.xAxis, runOn(this, this.zt), this.highColumn));
        }
        wa(a, b, c) {
            return this.we(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.yAxis, this.xAxis, runOn(this, this.i3), (d, e) => this.kb(d, e, this.xAxis, runOn(this, this.zt), this.lowColumn), (d, e) => this.j9(d, e, this.xAxis, runOn(this, this.zt), this.lowColumn));
        }
        wc(a, b, c) {
            return this.we(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.yAxis, this.xAxis, null, null, null);
        }
        zp() {
            return this.yAxis != null && this.yAxis.c1();
        }
        q9(a, b, c, d) {
            super.q9(a, b, c, d);
            switch (b) {
                case HorizontalRangeCategorySeries.$$p[0]:
                case HorizontalRangeCategorySeries.$$p[1]:
                case HorizontalRangeCategorySeries.$$p[2]:
                case HorizontalRangeCategorySeries.$$p[3]:
                    this.qf();
                    break;
                case HorizontalRangeCategorySeries.$$p[4]:
                    this.pc(typeCast(Axis.$, c));
                    this.rh(typeCast(Axis.$, d));
                    this.y3.cw.g(this.id);
                    this.rr(false);
                    this.ql();
                    break;
                case HorizontalRangeCategorySeries.$$p[5]:
                    this.pc(typeCast(Axis.$, c));
                    this.rh(typeCast(Axis.$, d));
                    this.y3.cw.g(this.id);
                    this.zp();
                    this.rr(false);
                    this.ql();
                    break;
            }
        }
        abj(a) {
            if (typeCast(NumericYAxis.$, a) !== null) {
                return true;
            }
            return false;
        }
        abi(a) {
            if (typeCast(CategoryXAxis.$, a) !== null || a.ce) {
                return true;
            }
            return false;
        }
        getDataLegendSeriesContextAt(a, b) {
            let c = { $type: Point_$type, x: a, y: b };
            let d = this.i3(c, false, false);
            let e = this.i1(c, false, false);
            let f = new DataLegendSeriesValueInfo();
            f.a = 1;
            f.l = e > d ? d : e;
            f.q = "Low";
            f.r = this.aa7 != null ? this.aa7 : "Low";
            f.s = this.abo != null ? this.abo : "L:";
            f.t = this.abq != null ? this.abq : "";
            let g = new DataLegendSeriesValueInfo();
            g.a = 2;
            g.l = e > d ? e : d;
            g.q = "High";
            g.r = this.aa4 != null ? this.aa4 : "High";
            g.s = this.abk != null ? this.abk : "H:";
            g.t = this.abm != null ? this.abm : "";
            let h = new DataLegendSeriesValueInfo();
            h.a = 7;
            h.l = Math.abs(g.l - f.l);
            h.q = "Range";
            h.r = "Range";
            h.s = "R:";
            h.k = true;
            let i = new DataLegendSeriesValueInfo();
            i.a = 6;
            i.l = f.l + (h.l / 2);
            i.q = "Average";
            i.r = "Average";
            i.s = "A:";
            i.k = true;
            let j = new DataLegendSeriesContext();
            j.e = ((() => {
                let $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
                $ret.add(g);
                $ret.add(f);
                $ret.add(h);
                $ret.add(i);
                return $ret;
            })());
            j.h = this.mc();
            return j;
        }
        get useHighMarkerFidelity() {
            return this.zq;
        }
    }
    HorizontalRangeCategorySeries.$t = /*@__PURE__*/ markType(HorizontalRangeCategorySeries, 'HorizontalRangeCategorySeries', RangeCategorySeries.$);
    HorizontalRangeCategorySeries.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, HorizontalRangeCategorySeries, 'raisePropertyChanged', ['HighMemberAsLegendLabel:abk:abu', [2, null], 'HighMemberAsLegendUnit:abm:abv', [2, null], 'LowMemberAsLegendLabel:abo:abw', [2, null], 'LowMemberAsLegendUnit:abq:abx', [2, null], 'XAxis::aby', [CategoryAxisBase.$, null], 'YAxis::abz', [NumericYAxis.$, null]]);
    return HorizontalRangeCategorySeries;
})();
