// igniteui-react-charts/esm2015/lib/igr-data-legend-series-group-info.js
import { DataLegendSeriesGroupInfo as DataLegendSeriesGroupInfo_internal } from "./DataLegendSeriesGroupInfo";
/**
 * Represent info about series values
*/
export class IgrDataLegendSeriesGroupInfo {
    createImplementation() {
        return new DataLegendSeriesGroupInfo_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
