// igniteui-react-charts/esm2015/lib/igr-radial-pie-series.js
import { LegendEmptyValuesMode_$type } from "igniteui-react-core";
import { IgrAnchoredRadialSeries } from "./igr-anchored-radial-series";
import { RadialPieSeries } from "./RadialPieSeries";
import { ensureEnum, toPoint, fromRect } from "igniteui-react-core";
/**
 * Represents a IgxDataChartComponent radial pie series.
 * Compare values across categories by using radial pie slices.
*/
export class IgrRadialPieSeries extends IgrAnchoredRadialSeries {
    createImplementation() {
        return new RadialPieSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the current series shows a pie shape.
    */
    get isPie() {
        return this.i.fn;
    }
    /**
     * Overridden by derived series classes to indicate when marker-less display is preferred or not.
    */
    get isMarkerlessDisplayPreferred() {
        return this.i.fl;
    }
    /**
     * Gets or sets the x-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusX() {
        return this.i.adv;
    }
    set radiusX(v) {
        this.i.adv = +v;
    }
    /**
     * Gets or sets the y-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusY() {
        return this.i.adw;
    }
    set radiusY(v) {
        this.i.adw = +v;
    }
    /**
     * Gets or sets the property name that contains the legend labels.
    */
    get legendLabelMemberPath() {
        return this.i.ad0;
    }
    set legendLabelMemberPath(v) {
        this.i.ad0 = v;
    }
    /**
     * Determines display mode for zero values in the legend.
    */
    get legendEmptyValuesMode() {
        return this.i.adq;
    }
    set legendEmptyValuesMode(v) {
        this.i.adq = ensureEnum(LegendEmptyValuesMode_$type, v);
    }
    /**
     * For a category plotted series, returns the current width of the items within the categories. This only returns a value if the items have some form of width (e.g. columns, bars, etc.) otherwise 0 is returned.
    
    */
    getItemSpan() {
        let iv = this.i.iw();
        return (iv);
    }
    getSeriesValueBoundingBox(world) {
        let iv = this.i.wn(toPoint(world));
        return fromRect(iv);
    }
}
