// igniteui-react-charts/esm2015/lib/igr-scaler-params.js
/**
 * Class containing several properties which are used as parameters passed to scaling operations in a SeriesViewer.
*
* Class containing several properties which are used as parameters passed to scaling operations in a `SeriesViewer`.
*/
export class IgrScalerParams {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * The series reference value.
     * Typically, the reference value is the initial value of the series.  The reference value is used by the PercentChangeYAxis, where all values in the series are scaled relative to the initial value.
    */
    get referenceValue() {
        return this.i.b;
    }
    set referenceValue(v) {
        this.i.b = +v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
