// igniteui-react-charts/esm2015/lib/igr-refresh-completed-event-args.js
import { RefreshCompletedEventArgs as RefreshCompletedEventArgs_internal } from "./RefreshCompletedEventArgs";
export class IgrRefreshCompletedEventArgs {
    createImplementation() {
        return new RefreshCompletedEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
}
