// igniteui-react-charts/esm2015/lib/igr-series-layer-property-overlay-value-resolving-event-args.js
import { SeriesLayerPropertyOverlayValueResolvingEventArgs as SeriesLayerPropertyOverlayValueResolvingEventArgs_internal } from "./SeriesLayerPropertyOverlayValueResolvingEventArgs";
/**
 * Provides data for Axis RangeChanged events.
*/
export class IgrSeriesLayerPropertyOverlayValueResolvingEventArgs {
    createImplementation() {
        return new SeriesLayerPropertyOverlayValueResolvingEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get value() {
        return this.i.value;
    }
    set value(v) {
        this.i.value = v;
    }
}
