// igniteui-react-charts/esm2015/lib/igr-step-line-series.js
import { IgrHorizontalAnchoredCategorySeries } from "./igr-horizontal-anchored-category-series";
import { StepLineSeries } from "./StepLineSeries";
/**
 * Represents a IgxDataChartComponent step line series.
*
* ```ts
* <IgrDataChart
*     dataSource={this.state.dataSource}
*     width="700px"
*     height="500px">
*
*     <IgrCategoryXAxis name="xAxis" label="Year" />
*     <IgrNumericYAxis  name="yAxis" />
*
*     <IgrStepLineSeries
*         name="series1"
*         xAxisName="xAxis"
*         yAxisName="yAxis"
* 		valueMemberPath="value" />
*  </IgrDataChart>
* ```
*
* ```ts
*   let series = new IgrStepLineSeries({name:"series1"});
*   series.xAxis = this.xAxis;
*   series.yAxis = this.yAxis;
*   series.valueMemberPath="value";
*   this.chart.series.add(series);
* ```
*/
export class IgrStepLineSeries extends IgrHorizontalAnchoredCategorySeries {
    createImplementation() {
        return new StepLineSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the current series shows only line shapes.
    */
    get isLineOnly() {
        return this.i.fk;
    }
    /**
     * Gets whether the current series shows step shapes.
    */
    get isStep() {
        return this.i.f1;
    }
    /**
     * Overridden by derived series classes to indicate when marker-less display is preferred or not.
    */
    get isMarkerlessDisplayPreferred() {
        return this.i.fl;
    }
    /**
     * Gets whether the current series shows an area or line shape.
    */
    get isAreaOrLine() {
        return this.i.ev;
    }
}
