// igniteui-react-charts/esm2015/lib/igr-pie-slice-others-context.js
import { PieSliceOthersContext as PieSliceOthersContext_internal } from "./PieSliceOthersContext";
/**
 * Represents the selected data contained inside the Others slice.
*/
export class IgrPieSliceOthersContext {
    createImplementation() {
        return new PieSliceOthersContext_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * The underlying data contained by the Others pie slice.
    */
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
