// igniteui-react-charts/esm2015/lib/igr-legend-mouse-event-args.js
/**
 * Provides data for legend mouse button related events.
*/
export class IgrLegendMouseEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * Gets a reference to the object that raised the event.
    */
    get originalSource() {
        return this.i.originalSource;
    }
    /**
     * Gets the ItemsSource item associated with the current event.
    */
    get item() {
        return this.i.item;
    }
    set item(v) {
        this.i.item = v;
    }
    /**
     * Gets the series associated with the current event.
    */
    get series() {
        return this.i.series;
    }
    set series(v) {
        this.i.series = v;
    }
    /**
     * Gets the Chart associated with the current event.
    */
    get chart() {
        return this.i.chart;
    }
    set chart(v) {
        this.i.chart = v;
    }
    /**
     * The legend item in context.
    */
    get legendItem() {
        return this.i.legendItem;
    }
    set legendItem(v) {
        this.i.legendItem = v;
    }
    /**
     * Provides a human readable description of the mouse button event.
    
    */
    toString() {
        let iv = this.i.toString();
        return (iv);
    }
}
