// igniteui-react-charts/esm2015/lib/igr-chaikin-oscillator-indicator.js
import { IgrStrategyBasedIndicator } from "./igr-strategy-based-indicator";
import { ChaikinOscillatorIndicator } from "./ChaikinOscillatorIndicator";
/**
 * Represents the IgxDataChartComponent Chaikin Oscillator indicator series.
 * Default required members: Close, Low, High, Volume
*
* `ChaikinOscillatorIndicator` class specify the series as Chaikin Oscillator Indicator series.
*
* ```ts
* <IgrDataChart
* 	ref={this.onChartRef}
*     dataSource={this.state.dataSource}
*     width="700px"
*     height="500px">
*
*
*     <IgrCategoryXAxis name="xAxis" label="Date" />
*     <IgrNumericYAxis  name="yAxis" />
*
*    <IgrChaikinOscillatorIndicator
*     name="series3"
*     xAxisName="xAxis"
*     yAxisName="yAxis"
*     displayType="Line"
*     lowMemberPath="Low"
*     highMemberPath="High"
*     openMemberPath="Open"
*     closeMemberPath="Close"
*     volumeMemberPath="Volume"
* 	longPeriod={30}/>
*  </IgrDataChart>
* ```
*
* ```ts
* let series1 = new IgrAbsoluteVolumeOscillatorIndicator({ name: ""});
* series1.xAxis = this.categoryXAxis;
* series1.yAxis = this.numericYAxis;
* series1.xAxisName = "XAxis";
* series1.yAxisName = "YAxis";
* series1.valueMemberPath = "Volume";
* series1.openMemberPath = "open";
* series1.highMemberPath = "high";
* series1.lowMemberPath = "low";
* series1.closeMemberPath = "close";
* ```
*/
export class IgrChaikinOscillatorIndicator extends IgrStrategyBasedIndicator {
    createImplementation() {
        return new ChaikinOscillatorIndicator();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or sets the short moving average period for the current ChaikinOscillatorIndicator object.
     * The typical, and initial, value for short CHO periods is 3.
    *
    * Set the `ShortPeriod` property if you wish to change the short moving average period for the indicator.
    *
    * ```ts
    * this.series.shortPeriod = 30;
    * ```
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *    <IgrChaikinOscillatorIndicator
    *     name="series3"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Line"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume"
    * 	ShortPeriod ={5}/>
    *  </IgrDataChart>
    * ```
    */
    get shortPeriod() {
        return this.i.shortPeriod;
    }
    set shortPeriod(v) {
        this.i.shortPeriod = +v;
    }
    /**
     * Gets or sets the long moving average period for the current ChaikinOscillatorIndicator object.
     * The typical, and initial, value for long CHO periods is 10.
    *
    * Set the `LongPeriod` property if you wish to change the long moving average period for the indicator.
    *
    * ```ts
    * this.series.longPeriod = 30;
    * ```
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *
    *     <IgrCategoryXAxis name="xAxis" label="Date" />
    *     <IgrNumericYAxis  name="yAxis" />
    *
    *    <IgrChaikinOscillatorIndicator
    *     name="series3"
    *     xAxisName="xAxis"
    *     yAxisName="yAxis"
    *     displayType="Line"
    *     lowMemberPath="Low"
    *     highMemberPath="High"
    *     openMemberPath="Open"
    *     closeMemberPath="Close"
    *     volumeMemberPath="Volume"
    * 	longPeriod={30}/>
    *  </IgrDataChart>
    * ```
    */
    get longPeriod() {
        return this.i.longPeriod;
    }
    set longPeriod(v) {
        this.i.longPeriod = +v;
    }
}
