// igniteui-react-charts/esm2015/lib/igr-stacked-spline-area-series.js
import { IgrHorizontalStackedSeriesBase } from "./igr-horizontal-stacked-series-base";
import { StackedSplineAreaSeries } from "./StackedSplineAreaSeries";
import { ensureBool } from "igniteui-react-core";
/**
 * Represents a IgxDataChartComponent stacked spline area series.
*/
export class IgrStackedSplineAreaSeries extends IgrHorizontalStackedSeriesBase {
    createImplementation() {
        return new StackedSplineAreaSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or sets whether to include the spline shape in the axis range requested of the axis.
    */
    get isSplineShapePartOfRange() {
        return this.i.ab3;
    }
    set isSplineShapePartOfRange(v) {
        this.i.ab3 = ensureBool(v);
    }
    /**
     * Gets whether the current series shows an area shape.
    */
    get isArea() {
        return this.i.eu;
    }
    /**
     * Gets whether the current series shows a spline shape.
    */
    get isSpline() {
        return this.i.f0;
    }
}
