// igniteui-react-charts/esm2015/lib/igr-domain-chart.js
import * as React from 'react';
import { HorizontalAlignment_$type } from "igniteui-react-core";
import { UnknownValuePlotting_$type } from "igniteui-react-core";
import { TrendLineType_$type } from "igniteui-react-core";
import { IgrChartSeriesEventArgs } from './igr-chart-series-event-args';
import { arrayFindByName, brushToString, stringToBrush, toBrushCollection, fromBrushCollection, fromRect, toRect, ensureBool, ensureEnum, isValidProp, getModifiedProps, toSpinal, initializePropertiesFromCss, NamePatcher } from "igniteui-react-core";
import { MarkerType_$type } from "./MarkerType";
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { IgrDomainChartSeriesPointerEventArgs } from './igr-domain-chart-series-pointer-event-args';
import { MarkerTypeCollection as MarkerTypeCollection_internal } from './MarkerTypeCollection';
import { IgrMarkerTypeCollection } from './igr-marker-type-collection';
import { SyncableObservableCollection$1 } from "igniteui-react-core";
import { ToolTipType_$type } from "./ToolTipType";
import { CrosshairsDisplayMode_$type } from "./CrosshairsDisplayMode";
import { IgrCalloutStyleUpdatingEventArgs } from "./igr-callout-style-updating-event-args";
import { ComputedPlotAreaMarginMode_$type } from './ComputedPlotAreaMarginMode';
import { SeriesHighlightingMode_$type } from './SeriesHighlightingMode';
import { SeriesHighlightingBehavior_$type } from './SeriesHighlightingBehavior';
import { LegendHighlightingMode_$type } from './LegendHighlightingMode';
import { SeriesOutlineMode_$type } from './SeriesOutlineMode';
import { MarkerOutlineMode_$type } from './MarkerOutlineMode';
import { MarkerFillMode_$type } from './MarkerFillMode';
import { MarkerAutomaticBehavior_$type } from './MarkerAutomaticBehavior';
import { LegendItemBadgeMode_$type } from "igniteui-react-core";
import { LegendItemBadgeShape_$type } from "igniteui-react-core";
import { DataAbbreviationMode_$type } from "igniteui-react-core";
import { DataLegendValueMode_$type } from "igniteui-react-core";
import { DataLegendHeaderDateMode_$type } from "igniteui-react-core";
import { DataLegendHeaderTimeMode_$type } from "igniteui-react-core";
import { DataLegendUnitsMode_$type } from "igniteui-react-core";
import { DataLegendLabelMode_$type } from "igniteui-react-core";
import { DataLegendSummaryType_$type } from "igniteui-react-core";
import { Thickness } from "igniteui-react-core";
import { FontInfo } from "igniteui-react-core";
import { SeriesPlotAreaMarginHorizontalMode_$type } from './SeriesPlotAreaMarginHorizontalMode';
import { SeriesPlotAreaMarginVerticalMode_$type } from './SeriesPlotAreaMarginVerticalMode';
import { IgrCalloutPlacementPositionsCollection } from './igr-callout-placement-positions-collection';
import { CalloutPlacementPositions_$type } from './CalloutPlacementPositions';
import { SeriesViewerScrollbarMode_$type } from './SeriesViewerScrollbarMode';
import { SeriesViewerHorizontalScrollbarPosition_$type } from './SeriesViewerHorizontalScrollbarPosition';
import { SeriesViewerVerticalScrollbarPosition_$type } from './SeriesViewerVerticalScrollbarPosition';
import { IgrCalloutRenderStyleUpdatingEventArgs } from './igr-callout-render-style-updating-event-args';
import { CalloutPlacementPositionsCollection as CalloutPlacementPositionsCollection_internal } from './CalloutPlacementPositionsCollection';
import { DataToolTipLayerGroupingMode_$type } from "igniteui-react-core";
import { DataTooltipGroupedPositionX_$type } from "igniteui-react-core";
import { DataTooltipGroupedPositionY_$type } from "igniteui-react-core";
import { IgrChartSummaryDescriptionCollection } from './igr-chart-summary-description-collection';
import { IgrChartSortDescriptionCollection } from './igr-chart-sort-description-collection';
import { IgrFilterExpressionCollection } from "igniteui-react-core";
import { ValueModeCollection as ValueModeCollection_internal } from './ValueModeCollection';
import { IgrValueModeCollection } from './igr-value-mode-collection';
import { ValueLayerValueMode_$type } from './ValueLayerValueMode';
import { IgrFilterStringErrorsParsingEventArgs } from './igr-filter-string-errors-parsing-event-args';
import { SeriesHighlightedValuesDisplayMode_$type } from "igniteui-react-core";
import { SeriesSelectionMode_$type } from './SeriesSelectionMode';
import { SeriesSelectionBehavior_$type } from './SeriesSelectionBehavior';
import { ChartSelectedItemCollection as ChartSelectedItemCollection_internal } from "./ChartSelectedItemCollection";
import { ChartSelection } from "./ChartSelection";
import { IgrSeriesViewerSelectedSeriesItemsChangedEventArgs } from "./igr-series-viewer-selected-series-items-changed-event-args";
import { IgrChartSelectedItemCollection } from "./igr-chart-selected-item-collection";
import { CalloutCollisionMode_$type } from "igniteui-react-core";
import { Visibility_$type } from "igniteui-react-core";
import { IgrDomainChartPlotAreaPointerEventArgs } from './igr-domain-chart-plot-area-pointer-event-args';
export class IgrDomainChart extends React.Component {
    createImplementation() {
        return null;
    }
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor(props) {
        super(props);
        this._calloutsDataSource = null;
        this._selectedSeriesItems = null;
        this._focusedSeriesItems = null;
        this._markerTypes = null;
        this._calloutsAllowedPositions = null;
        this._valueLines = null;
        this.__p = null;
        this._hasUserValues = new Set();
        this._stylingContainer = null;
        this._stylingParent = null;
        this._inStyling = false;
        this._seriesAdded = null;
        this._seriesAdded_wrapped = null;
        this._seriesRemoved = null;
        this._seriesRemoved_wrapped = null;
        this._seriesPointerEnter = null;
        this._seriesPointerEnter_wrapped = null;
        this._seriesPointerLeave = null;
        this._seriesPointerLeave_wrapped = null;
        this._seriesPointerMove = null;
        this._seriesPointerMove_wrapped = null;
        this._seriesPointerDown = null;
        this._seriesPointerDown_wrapped = null;
        this._seriesPointerUp = null;
        this._seriesPointerUp_wrapped = null;
        this._seriesClick = null;
        this._seriesClick_wrapped = null;
        this._plotAreaPointerEnter = null;
        this._plotAreaPointerEnter_wrapped = null;
        this._plotAreaPointerLeave = null;
        this._plotAreaPointerLeave_wrapped = null;
        this._plotAreaPointerMove = null;
        this._plotAreaPointerMove_wrapped = null;
        this._plotAreaPointerDown = null;
        this._plotAreaPointerDown_wrapped = null;
        this._plotAreaPointerUp = null;
        this._plotAreaPointerUp_wrapped = null;
        this._calloutStyleUpdating = null;
        this._calloutStyleUpdating_wrapped = null;
        this._calloutRenderStyleUpdating = null;
        this._calloutRenderStyleUpdating_wrapped = null;
        this._selectedSeriesItemsChanged = null;
        this._selectedSeriesItemsChanged_wrapped = null;
        this._focusedSeriesItemsChanged = null;
        this._focusedSeriesItemsChanged_wrapped = null;
        this._filterStringErrorsParsing = null;
        this._filterStringErrorsParsing_wrapped = null;
        if (this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
        }
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    componentDidMount() {
        for (const p of Object.keys(this.props)) {
            if (isValidProp(this, p)) {
                this[p] = this.props[p];
            }
        }
    }
    shouldComponentUpdate(nextProps, nextState) {
        const mod = getModifiedProps(this.props, nextProps);
        for (const p of Object.keys(mod)) {
            if (isValidProp(this, p)) {
                this[p] = mod[p];
            }
        }
        return true;
    }
    render() {
        return null;
    }
    /**
    * Sets or gets the template reference to use for tooltips for all the series.
    *
    *
    */
    get tooltipTemplate() {
        return this.i.tooltipTemplate;
    }
    set tooltipTemplate(v) {
        this.i.tooltipTemplate = v;
    }
    /**
    * Sets or gets the template references to use for tooltips for all the series, in order.
    *
    *
    */
    get tooltipTemplates() {
        let ret = [];
        if (this.i.tooltipTemplates) {
            for (let ii = 0; ii < this.i.tooltipTemplates.count; ii++) {
                ret.push(this.i.tooltipTemplates.item(ii));
            }
        }
        return ret;
    }
    set tooltipTemplates(v) {
        this.i.tooltipTemplates.clear();
        if (v) {
            for (let ii = 0; ii < v.length; ii++) {
                this.i.tooltipTemplates.add(v[ii]);
            }
        }
    }
    get legend() {
        if (this.i.legend != null)
            return this.i.legend.externalObject;
    }
    set legend(v) {
        if (v != undefined && v != null)
            this.i.legend = v.i;
    }
    set calloutsDataSource(value) {
        this._calloutsDataSource = value;
        this.bindCalloutsData();
    }
    get calloutsDataSource() {
        return this._calloutsDataSource;
    }
    bindCalloutsData() {
        if (this.i != null && this.i !== undefined) {
            this.i.calloutsItemsSource = this._calloutsDataSource;
        }
    }
    /**
 * Gets or sets the scaling value used to affect the pixel density of the control.
 * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
 * to appear blurry.
*/
    get pixelScalingRatio() {
        return this.i.j0;
    }
    set pixelScalingRatio(v) {
        this.i.j0 = +v;
    }
    /**
     * Gets or sets the left margin of chart title
    *
    *  Use `titleLeftMargin` property for the left margin of chart title.
    *
    * ```ts
    * this.chart.titleLeftMargin = 10;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart dataSource={this.state.data}
    *                    titleLeftMargin={20}>
    *  </IgrCategoryChart>
    * ```
    */
    get titleLeftMargin() {
        return this.i.kd;
    }
    set titleLeftMargin(v) {
        this.i.kd = +v;
    }
    /**
     * Gets or sets the right margin of chart title
    *
    *  Use `titleLeftMargin` property for the right margin of chart title.
    *
    * ```ts
    * this.chart.titleRightMargin = 10;
    * ```
    *
    * ```ts
    * <IgrCategoryChart
    * 				dataSource={this.state.data}
    *                 titleRightMargin = 10>
    *  </IgrCategoryChart>
    * ```
    */
    get titleRightMargin() {
        return this.i.ke;
    }
    set titleRightMargin(v) {
        this.i.ke = +v;
    }
    /**
     * Gets or sets the top margin of chart title
    *
    *  Use `titleTopMargin` property for the top margin of chart title.
    *
    * ```ts
    * this.chart.titleTopMargin = 10;
    * ```
    *
    * ```ts
    * <IgrCategoryChart dataSource={this.state.data}
    *                titleTopMargin={10} />
    * ```
    */
    get titleTopMargin() {
        return this.i.kf;
    }
    set titleTopMargin(v) {
        this.i.kf = +v;
    }
    /**
     * Gets or sets the bottom margin of chart title
    *
    *  Use `titleBottomMargin` property for the bottom margin of chart title.
    *
    * ```ts
    *  this.chart.titleBottomMargin = 5;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	subtitle= "CategoryChart Subtitle"
    * 	subtitleAlignment="Right"
    * 	titleBottomMargin={5}/>
    * ```
    */
    get titleBottomMargin() {
        return this.i.kc;
    }
    set titleBottomMargin(v) {
        this.i.kc = +v;
    }
    /**
     * Gets or sets the left margin of chart subtitle
    *
    * Use `subtitleLeftMargin` property for the left margin of chart subtitle.
    *
    * ```ts
    * this.chart.subtitleLeftMargin = 20 ;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	subtitle= "CategoryChart Subtitle"
    * 	subtitleLeftMargin= {20}/>
    * ```
    */
    get subtitleLeftMargin() {
        return this.i.j8;
    }
    set subtitleLeftMargin(v) {
        this.i.j8 = +v;
    }
    /**
     * Gets or sets the top margin of chart subtitle
    *
    * Use `subtitleTopMargin` property for the top margin of chart subtitle.
    *
    * ```ts
    * this.chart.subtitleTopMargin = 10;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	subtitle= "CategoryChart Subtitle"
    * 	subtitleTopMargin= {10}/>
    * ```
    */
    get subtitleTopMargin() {
        return this.i.ka;
    }
    set subtitleTopMargin(v) {
        this.i.ka = +v;
    }
    /**
     * Gets or sets the right margin of chart subtitle
    *
    * Use `subtitleRightMargin` property for the right margin of chart subtitle.
    *
    * ```ts
    *  this.chart.subtitleRightMargin = 20 ;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	subtitle= "CategoryChart Subtitle"
    * 	subtitleAlignment="Right"
    * 	subtitleRightMargin ={20}/>
    * ```
    */
    get subtitleRightMargin() {
        return this.i.j9;
    }
    set subtitleRightMargin(v) {
        this.i.j9 = +v;
    }
    /**
     * Gets or sets the bottom margin of chart subtitle
    *
    * Use `subtitleBottomMargin` property for the bottom margin of chart subtitle.
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	subtitle= "CategoryChart Subtitle"
    * 	subtitleBottomMargin= {10}/>
    * ```
    *
    * ```ts
    * this.chart.subtitleBottomMargin = 10 ;
    * ```
    */
    get subtitleBottomMargin() {
        return this.i.j7;
    }
    set subtitleBottomMargin(v) {
        this.i.j7 = +v;
    }
    /**
     * Gets or sets color of chart subtitle
    *
    * Use `subtitleTextColor` property to color the subtitle.
    *
    * ```ts
    * this.chart.subtitleTextColor = "#ff0000" ;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	subtitle= "CategoryChart Subtitle"
    * 	subtitleTextColor="#ff0000"/>
    * ```
    */
    get subtitleTextColor() {
        return brushToString(this.i.yy);
    }
    set subtitleTextColor(v) {
        this.i.yy = stringToBrush(v);
    }
    /**
     * Gets or sets color of chart title
    *
    *  Use `titleTextColor` property to color the chart title
    *
    * ```ts
    *  this.chart.titleTextColor="red" ;
    * ```
    *
    * ```ts
    * <IgrCategoryChart
    * 		dataSource={this.state.data}
    * 		width="700px"
    * 		height="500px"
    * 		xAxisTitle="Countries"
    * 		titleTextColor="Red" />
    *
    * ```
    */
    get titleTextColor() {
        return brushToString(this.i.yz);
    }
    set titleTextColor(v) {
        this.i.yz = stringToBrush(v);
    }
    /**
     * Gets or sets the left margin of the chart content.
    *
    * Use the 'leftMargin' property for the left margin of the chart content.
    *
    * ```ts
    *  this.chart.leftMargin = 20;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart dataSource={this.state.data}
    *                    leftMargin= {20}/>
    * ```
    */
    get leftMargin() {
        return this.i.jx;
    }
    set leftMargin(v) {
        this.i.jx = +v;
    }
    /**
     * Gets or sets the top margin of the chart content.
    *
    * Use `topMargin` property for the margin of the chart content.
    *
    * ```ts
    * this.chart.topMargin=20;
    * ```
    *
    * ```ts
    * <IgrCategoryChart dataSource={this.state.data}
    *                topMargin={10} />
    * ```
    */
    get topMargin() {
        return this.i.kg;
    }
    set topMargin(v) {
        this.i.kg = +v;
    }
    /**
     * Gets or sets the right margin of the chart content.
    *
    * Use `rightMargin` property for the right margin of the chart content.
    *
    * ```ts
    *   this.chart.rightMargin = 20 ;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	markerTypes={this.state.markersTypes}
    * 	dataSource={this.categoryData}
    * 	rightMargin= {4}/>
    * ```
    */
    get rightMargin() {
        return this.i.j6;
    }
    set rightMargin(v) {
        this.i.j6 = +v;
    }
    /**
     * Gets or sets the bottom margin around the chart content.
    *
    * Use the `bottomMargin` property for the bottom margin around the chart content.
    *
    * ```ts
    * this.chart.bottomMargin=20;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart dataSource={this.state.data}
    *                    bottomMargin= 20 />
    * ```
    */
    get bottomMargin() {
        return this.i.jj;
    }
    set bottomMargin(v) {
        this.i.jj = +v;
    }
    /**
     * Gets or sets the duration used for animating series plots when the data is changing
    *
    * Use the `TransitionDuration` property to animating between data values.
    *
    * ```ts
    * this.chart.transitionDuratio= 500;
    * ```
    *
    * ```ts
    * <IgrCategoryChart dataSource={this.state.data}
    *                transitionDuratio={500} />
    * ```
    */
    get transitionDuration() {
        return this.i.lj;
    }
    set transitionDuration(v) {
        this.i.lj = +v;
    }
    /**
     * Gets or sets the easing function used for animating series plots when the data is changing.
     * This can be set to one of the known values "linear" or "cubic," or it can be set to an easing function which takes a single numeric parameter and returns a number.
    *
    * The 'transitionInEasingFunction' property used for easing function for animating series plots when the chart is loading into view.
    *
    * ```ts
    * this.chart.transitionInEasingFunction=  this.cubicFunc;
    *
    * cubicFunc(time: number) : number
    *   {
    *     return time;
    *   }
    * ```
    *
    * ```ts
    * <IgrCategoryChart dataSource={this.state.data}
    *                transitionInEasingFunction="cubic" />
    * ```
    */
    get transitionEasingFunction() {
        return this.i.es;
    }
    set transitionEasingFunction(v) {
        this.i.es = v;
    }
    /**
     * Gets or sets the duration used for animating highlighting changes
    */
    get highlightingTransitionDuration() {
        return this.i.lg;
    }
    set highlightingTransitionDuration(v) {
        this.i.lg = +v;
    }
    /**
     * Gets or sets the duration used for animating Selection changes
    */
    get selectionTransitionDuration() {
        return this.i.li;
    }
    set selectionTransitionDuration(v) {
        this.i.li = +v;
    }
    /**
     * Gets or sets the duration used for animating Focus changes
    */
    get focusTransitionDuration() {
        return this.i.lf;
    }
    set focusTransitionDuration(v) {
        this.i.lf = +v;
    }
    /**
     * Gets or sets CSS font property for the chart subtitle
    *
    * Use `subtitleTextStyle` property for the CSS font of the chart subtitle.
    *
    * ```ts
    * this.chart.subtitleTextStyle= "16pt Verdona";
    * ```
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	subtitleTextStyle="16pt Verdona"/>
    * ```
    */
    get subtitleTextStyle() {
        return this.i.sv;
    }
    set subtitleTextStyle(v) {
        this.i.sv = v;
    }
    /**
     * Gets or sets CSS font property for the chart title
    *
    * Use `titleTextStyle` property for the CSS font property of the chart title
    *
    * ```ts
    * this.chart.xAxisTitleTextStyle = "24pt Verdona";
    * ```
    *
    * ```ts
    * <IgrCategoryChart dataSource={this.state.data}
    *                 width="700px"
    *                 height="500px"
    *                 xAxisTitle="Countries"
    *                 xAxisTitleTextColor="gray"
    *                 xAxisTitleTextStyle="12pt Verdana"
    *                 xAxisTitleAngle={0}
    *                 yAxisTitle="Trillions of Watt-hours (TWh)"
    *                 yAxisTitleTextStyle="12pt Verdana"
    *                 yAxisTitleTextColor="gray"
    *                 yAxisTitleAngle={90}
    *                 yAxisTitleLeftMargin={5}>
    *  </IgrCategoryChart>
    * ```
    */
    get titleTextStyle() {
        return this.i.s6;
    }
    set titleTextStyle(v) {
        this.i.s6 = v;
    }
    get isDetached() {
        return this.i.isDetached;
    }
    get dataToolTipGroupingMode() {
        return this.i.bz;
    }
    set dataToolTipGroupingMode(v) {
        this.i.bz = ensureEnum(DataToolTipLayerGroupingMode_$type, v);
    }
    /**
     * Gets or sets the offset of the tooltip layer on the X axis.
    */
    get dataToolTipPositionOffsetX() {
        return this.i.jo;
    }
    set dataToolTipPositionOffsetX(v) {
        this.i.jo = +v;
    }
    /**
     * Gets or sets the offset of the tooltip layer on the Y axis.
    */
    get dataToolTipPositionOffsetY() {
        return this.i.jp;
    }
    set dataToolTipPositionOffsetY(v) {
        this.i.jp = +v;
    }
    /**
     * Gets or sets the offset of the tooltip layer on the X axis.
    */
    get dataToolTipDefaultPositionOffsetX() {
        return this.i.jm;
    }
    set dataToolTipDefaultPositionOffsetX(v) {
        this.i.jm = +v;
    }
    /**
     * Gets or sets the offset of the tooltip layer on the Y axis.
    */
    get dataToolTipDefaultPositionOffsetY() {
        return this.i.jn;
    }
    set dataToolTipDefaultPositionOffsetY(v) {
        this.i.jn = +v;
    }
    /**
     * Gets or sets the tooltip position mode on the X axis for grouped series.
    */
    get dataToolTipGroupedPositionModeX() {
        return this.i.bv;
    }
    set dataToolTipGroupedPositionModeX(v) {
        this.i.bv = ensureEnum(DataTooltipGroupedPositionX_$type, v);
    }
    /**
     * Gets or sets the tooltip position mode on the Y axis for grouped series.
    */
    get dataToolTipGroupedPositionModeY() {
        return this.i.bx;
    }
    set dataToolTipGroupedPositionModeY(v) {
        this.i.bx = ensureEnum(DataTooltipGroupedPositionY_$type, v);
    }
    /**
     * Gets or sets whether the data legend should update when the series data is mutated.
    */
    get dataToolTipShouldUpdateWhenSeriesDataChanges() {
        return this.i.gq;
    }
    set dataToolTipShouldUpdateWhenSeriesDataChanges(v) {
        this.i.gq = ensureBool(v);
    }
    /**
     * Gets or sets indexes, titles, or names of series to include in displaying in the data tooltip, e.g. "0, 1" or "Series1 Title, Series2 Title"
     * The DataToolTipExcludedSeries property takes precedence over values of DataToolTipIncludedSeries property
    */
    get dataToolTipIncludedSeries() {
        return this.i.m;
    }
    set dataToolTipIncludedSeries(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.m = v;
    }
    /**
     * Gets or sets indexes, titles, or names of series to exclude from displaying in the data tooltip, e.g. "0, 1" or "Series1 Title, Series2 Title"
     * The DataToolTipExcludedSeries property takes precedence over values of DataToolTipIncludedSeries property
    */
    get dataToolTipExcludedSeries() {
        return this.i.j;
    }
    set dataToolTipExcludedSeries(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.j = v;
    }
    /**
     * Gets or sets names of data columns or their labels to include in displaying in the data legend, e.g. "High, Low" or "H, L"
     * The DataToolTipExcludedColumns property takes precedence over values of DataToolTipIncludedColumns property
    */
    get dataToolTipIncludedColumns() {
        return this.i.l;
    }
    set dataToolTipIncludedColumns(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.l = v;
    }
    /**
     * Gets or sets names of data columns or their labels to exclude from displaying in the data legend, e.g. "High, Low" or "H, L"
     * The DataToolTipExcludedColumns property takes precedence over values of DataToolTipIncludedColumns property
    */
    get dataToolTipExcludedColumns() {
        return this.i.i;
    }
    set dataToolTipExcludedColumns(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.i = v;
    }
    /**
     * Gets or sets mode for abbreviating large numbers displayed in the legend
     * This property is ignored when setting ValueFormatSpecifiers or ValueFormatString properties
    */
    get dataToolTipValueFormatAbbreviation() {
        return this.i.bd;
    }
    set dataToolTipValueFormatAbbreviation(v) {
        this.i.bd = ensureEnum(DataAbbreviationMode_$type, v);
    }
    /**
     * Gets or sets maximum digits for formatting numbers displayed in the legend
     * This property is ignored when setting ValueFormatSpecifiers or ValueFormatString properties
    */
    get dataToolTipValueFormatMaxFractions() {
        return this.i.lc;
    }
    set dataToolTipValueFormatMaxFractions(v) {
        this.i.lc = +v;
    }
    /**
     * Gets or sets minimum digits for formatting numbers displayed in the legend
     * This property is ignored when setting ValueFormatSpecifiers or ValueFormatString properties
    */
    get dataToolTipValueFormatMinFractions() {
        return this.i.ld;
    }
    set dataToolTipValueFormatMinFractions(v) {
        this.i.ld = +v;
    }
    /**
     * Gets or sets the mode for displaying values in the data legend, e.g. Currency ($500.25), Decimal (500.25), Integer (500)
    */
    get dataToolTipValueFormatMode() {
        return this.i.bt;
    }
    set dataToolTipValueFormatMode(v) {
        this.i.bt = ensureEnum(DataLegendValueMode_$type, v);
    }
    /**
     * Gets or sets globalization culture when displaying values as currencies, e.g. use "en-GB" to display British pound symbol when the ValueFormatMode property is set to 'Currency' mode
     * This property is ignored when setting ValueFormatSpecifiers or ValueFormatString properties
    */
    get dataToolTipValueFormatCulture() {
        return this.i.py;
    }
    set dataToolTipValueFormatCulture(v) {
        this.i.py = v;
    }
    /**
     * Gets or sets whether or not use grouping separator, e.g, 15,000 for 15000
     * This property is ignored when setting ValueFormatSpecifiers or ValueFormatString properties
    */
    get dataToolTipValueFormatUseGrouping() {
        return this.i.gr;
    }
    set dataToolTipValueFormatUseGrouping(v) {
        this.i.gr = ensureBool(v);
    }
    /**
     * Gets or sets the format string for values displayed in the data legend.
    */
    get dataToolTipValueFormatString() {
        return this.i.p4;
    }
    set dataToolTipValueFormatString(v) {
        this.i.p4 = v;
    }
    /**
     * Gets or sets the format specifiers to use with the ValueFormatString string.
    */
    get dataToolTipValueFormatSpecifiers() {
        return this.i.n;
    }
    set dataToolTipValueFormatSpecifiers(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.n = v;
    }
    get dataToolTipValueRowMarginBottom() {
        return this.i.zs ? this.i.zs.bottom : NaN;
    }
    set dataToolTipValueRowMarginBottom(v) {
        this.ensureDataToolTipValueRowMargin();
        this.i.zs.bottom = +v;
        this.i.zs = this.i.zs;
    }
    get dataToolTipValueRowMarginLeft() {
        return this.i.zs ? this.i.zs.left : NaN;
    }
    set dataToolTipValueRowMarginLeft(v) {
        this.ensureDataToolTipValueRowMargin();
        this.i.zs.left = +v;
        this.i.zs = this.i.zs;
    }
    get dataToolTipValueRowMarginRight() {
        return this.i.zs ? this.i.zs.right : NaN;
    }
    set dataToolTipValueRowMarginRight(v) {
        this.ensureDataToolTipValueRowMargin();
        this.i.zs.right = +v;
        this.i.zs = this.i.zs;
    }
    get dataToolTipValueRowMarginTop() {
        return this.i.zs ? this.i.zs.top : NaN;
    }
    set dataToolTipValueRowMarginTop(v) {
        this.ensureDataToolTipValueRowMargin();
        this.i.zs.top = +v;
        this.i.zs = this.i.zs;
    }
    ensureDataToolTipValueRowMargin() {
        if (this.i.zs) {
            return;
        }
        this.i.zs = new Thickness(2);
    }
    /**
     * Gets or sets whether to show series rows.
    */
    get dataToolTipValueRowVisible() {
        return this.i.gs;
    }
    set dataToolTipValueRowVisible(v) {
        this.i.gs = ensureBool(v);
    }
    /**
     * Gets or sets text displayed when data column is missing a value, e.g. "no data"
    */
    get dataToolTipValueTextWhenMissingData() {
        return this.i.qd;
    }
    set dataToolTipValueTextWhenMissingData(v) {
        this.i.qd = v;
    }
    /**
     * Gets or sets whether to use series colors when displaying values in the legend
    */
    get dataToolTipValueTextUseSeriesColors() {
        return this.i.gt;
    }
    set dataToolTipValueTextUseSeriesColors(v) {
        this.i.gt = ensureBool(v);
    }
    get dataToolTipValueTextMarginBottom() {
        return this.i.zt ? this.i.zt.bottom : NaN;
    }
    set dataToolTipValueTextMarginBottom(v) {
        this.ensureDataToolTipValueTextMargin();
        this.i.zt.bottom = +v;
        this.i.zt = this.i.zt;
    }
    get dataToolTipValueTextMarginLeft() {
        return this.i.zt ? this.i.zt.left : NaN;
    }
    set dataToolTipValueTextMarginLeft(v) {
        this.ensureDataToolTipValueTextMargin();
        this.i.zt.left = +v;
        this.i.zt = this.i.zt;
    }
    get dataToolTipValueTextMarginRight() {
        return this.i.zt ? this.i.zt.right : NaN;
    }
    set dataToolTipValueTextMarginRight(v) {
        this.ensureDataToolTipValueTextMargin();
        this.i.zt.right = +v;
        this.i.zt = this.i.zt;
    }
    get dataToolTipValueTextMarginTop() {
        return this.i.zt ? this.i.zt.top : NaN;
    }
    set dataToolTipValueTextMarginTop(v) {
        this.ensureDataToolTipValueTextMargin();
        this.i.zt.top = +v;
        this.i.zt = this.i.zt;
    }
    ensureDataToolTipValueTextMargin() {
        if (this.i.zt) {
            return;
        }
        this.i.zt = new Thickness(2);
    }
    /**
     * Gets or sets the units text color.
    */
    get dataToolTipValueTextColor() {
        return brushToString(this.i.ym);
    }
    set dataToolTipValueTextColor(v) {
        this.i.ym = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get dataToolTipValueTextStyle() {
        if (this.i.fe == null) {
            return null;
        }
        return this.i.fe.fontString;
    }
    set dataToolTipValueTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.fe = fi;
    }
    /**
     * Gets or sets the format string for header text displayed in the data legend.
    */
    get dataToolTipHeaderFormatString() {
        return this.i.oo;
    }
    set dataToolTipHeaderFormatString(v) {
        this.i.oo = v;
    }
    /**
     * Gets or sets the format specifiers to use with the HeaderFormatString string.
    */
    get dataToolTipHeaderFormatSpecifiers() {
        return this.i.k;
    }
    set dataToolTipHeaderFormatSpecifiers(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.k = v;
    }
    /**
     * Gets or sets globalization culture when displaying header as date time
     * This property is ignored when setting HeaderFormatSpecifiers or HeaderFormatString properties
    */
    get dataToolTipHeaderFormatCulture() {
        return this.i.ok;
    }
    set dataToolTipHeaderFormatCulture(v) {
        this.i.ok = v;
    }
    /**
     * Gets or sets date format for the header
     * This property is ignored when setting HeaderFormatSpecifiers or HeaderFormatString properties
    */
    get dataToolTipHeaderFormatDate() {
        return this.i.bj;
    }
    set dataToolTipHeaderFormatDate(v) {
        this.i.bj = ensureEnum(DataLegendHeaderDateMode_$type, v);
    }
    /**
     * Gets or sets time format for the header
     * This property is ignored when setting HeaderFormatSpecifiers or HeaderFormatString properties
    */
    get dataToolTipHeaderFormatTime() {
        return this.i.bl;
    }
    set dataToolTipHeaderFormatTime(v) {
        this.i.bl = ensureEnum(DataLegendHeaderTimeMode_$type, v);
    }
    /**
     * Gets or sets the HeaderText for the data legend.
    */
    get dataToolTipHeaderText() {
        return this.i.ot;
    }
    set dataToolTipHeaderText(v) {
        this.i.ot = v;
    }
    /**
     * Gets or sets the header text color.
    */
    get dataToolTipHeaderTextColor() {
        return brushToString(this.i.ye);
    }
    set dataToolTipHeaderTextColor(v) {
        this.i.ye = stringToBrush(v);
    }
    get dataToolTipHeaderTextMarginBottom() {
        return this.i.zm ? this.i.zm.bottom : NaN;
    }
    set dataToolTipHeaderTextMarginBottom(v) {
        this.ensureDataToolTipHeaderTextMargin();
        this.i.zm.bottom = +v;
        this.i.zm = this.i.zm;
    }
    get dataToolTipHeaderTextMarginLeft() {
        return this.i.zm ? this.i.zm.left : NaN;
    }
    set dataToolTipHeaderTextMarginLeft(v) {
        this.ensureDataToolTipHeaderTextMargin();
        this.i.zm.left = +v;
        this.i.zm = this.i.zm;
    }
    get dataToolTipHeaderTextMarginRight() {
        return this.i.zm ? this.i.zm.right : NaN;
    }
    set dataToolTipHeaderTextMarginRight(v) {
        this.ensureDataToolTipHeaderTextMargin();
        this.i.zm.right = +v;
        this.i.zm = this.i.zm;
    }
    get dataToolTipHeaderTextMarginTop() {
        return this.i.zm ? this.i.zm.top : NaN;
    }
    set dataToolTipHeaderTextMarginTop(v) {
        this.ensureDataToolTipHeaderTextMargin();
        this.i.zm.top = +v;
        this.i.zm = this.i.zm;
    }
    ensureDataToolTipHeaderTextMargin() {
        if (this.i.zm) {
            return;
        }
        this.i.zm = new Thickness(2);
    }
    get dataToolTipHeaderRowMarginBottom() {
        return this.i.zl ? this.i.zl.bottom : NaN;
    }
    set dataToolTipHeaderRowMarginBottom(v) {
        this.ensureDataToolTipHeaderRowMargin();
        this.i.zl.bottom = +v;
        this.i.zl = this.i.zl;
    }
    get dataToolTipHeaderRowMarginLeft() {
        return this.i.zl ? this.i.zl.left : NaN;
    }
    set dataToolTipHeaderRowMarginLeft(v) {
        this.ensureDataToolTipHeaderRowMargin();
        this.i.zl.left = +v;
        this.i.zl = this.i.zl;
    }
    get dataToolTipHeaderRowMarginRight() {
        return this.i.zl ? this.i.zl.right : NaN;
    }
    set dataToolTipHeaderRowMarginRight(v) {
        this.ensureDataToolTipHeaderRowMargin();
        this.i.zl.right = +v;
        this.i.zl = this.i.zl;
    }
    get dataToolTipHeaderRowMarginTop() {
        return this.i.zl ? this.i.zl.top : NaN;
    }
    set dataToolTipHeaderRowMarginTop(v) {
        this.ensureDataToolTipHeaderRowMargin();
        this.i.zl.top = +v;
        this.i.zl = this.i.zl;
    }
    ensureDataToolTipHeaderRowMargin() {
        if (this.i.zl) {
            return;
        }
        this.i.zl = new Thickness(2);
    }
    /**
     * Gets or sets whether to show Header row.
    */
    get dataToolTipHeaderRowVisible() {
        return this.i.gp;
    }
    set dataToolTipHeaderRowVisible(v) {
        this.i.gp = ensureBool(v);
    }
    /**
     * Gets or Sets the style to use for the header text.
    */
    get dataToolTipHeaderTextStyle() {
        if (this.i.e6 == null) {
            return null;
        }
        return this.i.e6.fontString;
    }
    set dataToolTipHeaderTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.e6 = fi;
    }
    /**
     * Gets or sets the Group text color.
    */
    get dataToolTipGroupTextColor() {
        return brushToString(this.i.yd);
    }
    set dataToolTipGroupTextColor(v) {
        this.i.yd = stringToBrush(v);
    }
    get dataToolTipGroupTextMarginBottom() {
        return this.i.zk ? this.i.zk.bottom : NaN;
    }
    set dataToolTipGroupTextMarginBottom(v) {
        this.ensureDataToolTipGroupTextMargin();
        this.i.zk.bottom = +v;
        this.i.zk = this.i.zk;
    }
    get dataToolTipGroupTextMarginLeft() {
        return this.i.zk ? this.i.zk.left : NaN;
    }
    set dataToolTipGroupTextMarginLeft(v) {
        this.ensureDataToolTipGroupTextMargin();
        this.i.zk.left = +v;
        this.i.zk = this.i.zk;
    }
    get dataToolTipGroupTextMarginRight() {
        return this.i.zk ? this.i.zk.right : NaN;
    }
    set dataToolTipGroupTextMarginRight(v) {
        this.ensureDataToolTipGroupTextMargin();
        this.i.zk.right = +v;
        this.i.zk = this.i.zk;
    }
    get dataToolTipGroupTextMarginTop() {
        return this.i.zk ? this.i.zk.top : NaN;
    }
    set dataToolTipGroupTextMarginTop(v) {
        this.ensureDataToolTipGroupTextMargin();
        this.i.zk.top = +v;
        this.i.zk = this.i.zk;
    }
    ensureDataToolTipGroupTextMargin() {
        if (this.i.zk) {
            return;
        }
        this.i.zk = new Thickness(2);
    }
    get dataToolTipGroupRowMarginBottom() {
        return this.i.zj ? this.i.zj.bottom : NaN;
    }
    set dataToolTipGroupRowMarginBottom(v) {
        this.ensureDataToolTipGroupRowMargin();
        this.i.zj.bottom = +v;
        this.i.zj = this.i.zj;
    }
    get dataToolTipGroupRowMarginLeft() {
        return this.i.zj ? this.i.zj.left : NaN;
    }
    set dataToolTipGroupRowMarginLeft(v) {
        this.ensureDataToolTipGroupRowMargin();
        this.i.zj.left = +v;
        this.i.zj = this.i.zj;
    }
    get dataToolTipGroupRowMarginRight() {
        return this.i.zj ? this.i.zj.right : NaN;
    }
    set dataToolTipGroupRowMarginRight(v) {
        this.ensureDataToolTipGroupRowMargin();
        this.i.zj.right = +v;
        this.i.zj = this.i.zj;
    }
    get dataToolTipGroupRowMarginTop() {
        return this.i.zj ? this.i.zj.top : NaN;
    }
    set dataToolTipGroupRowMarginTop(v) {
        this.ensureDataToolTipGroupRowMargin();
        this.i.zj.top = +v;
        this.i.zj = this.i.zj;
    }
    ensureDataToolTipGroupRowMargin() {
        if (this.i.zj) {
            return;
        }
        this.i.zj = new Thickness(2);
    }
    /**
     * Gets or sets whether to show Group row.
    */
    get dataToolTipGroupRowVisible() {
        return this.i.go;
    }
    set dataToolTipGroupRowVisible(v) {
        this.i.go = ensureBool(v);
    }
    /**
     * Gets or Sets the style to use for the Group text.
    */
    get dataToolTipGroupTextStyle() {
        if (this.i.e5 == null) {
            return null;
        }
        return this.i.e5.fontString;
    }
    set dataToolTipGroupTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.e5 = fi;
    }
    /**
     * Gets or sets the summary text color.
    */
    get dataToolTipSummaryTitleTextColor() {
        return brushToString(this.i.yh);
    }
    set dataToolTipSummaryTitleTextColor(v) {
        this.i.yh = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the summary text.
    */
    get dataToolTipSummaryTitleTextStyle() {
        if (this.i.e9 == null) {
            return null;
        }
        return this.i.e9.fontString;
    }
    set dataToolTipSummaryTitleTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.e9 = fi;
    }
    /**
     * Gets or sets the SummaryType for the data legend.
    */
    get dataToolTipSummaryType() {
        return this.i.bp;
    }
    set dataToolTipSummaryType(v) {
        this.i.bp = ensureEnum(DataLegendSummaryType_$type, v);
    }
    /**
     * Gets or sets the SummaryTitleText for the data legend.
    */
    get dataToolTipSummaryTitleText() {
        return this.i.pc;
    }
    set dataToolTipSummaryTitleText(v) {
        this.i.pc = v;
    }
    get dataToolTipSummaryTitleTextMarginBottom() {
        return this.i.zp ? this.i.zp.bottom : NaN;
    }
    set dataToolTipSummaryTitleTextMarginBottom(v) {
        this.ensureDataToolTipSummaryTitleTextMargin();
        this.i.zp.bottom = +v;
        this.i.zp = this.i.zp;
    }
    get dataToolTipSummaryTitleTextMarginLeft() {
        return this.i.zp ? this.i.zp.left : NaN;
    }
    set dataToolTipSummaryTitleTextMarginLeft(v) {
        this.ensureDataToolTipSummaryTitleTextMargin();
        this.i.zp.left = +v;
        this.i.zp = this.i.zp;
    }
    get dataToolTipSummaryTitleTextMarginRight() {
        return this.i.zp ? this.i.zp.right : NaN;
    }
    set dataToolTipSummaryTitleTextMarginRight(v) {
        this.ensureDataToolTipSummaryTitleTextMargin();
        this.i.zp.right = +v;
        this.i.zp = this.i.zp;
    }
    get dataToolTipSummaryTitleTextMarginTop() {
        return this.i.zp ? this.i.zp.top : NaN;
    }
    set dataToolTipSummaryTitleTextMarginTop(v) {
        this.ensureDataToolTipSummaryTitleTextMargin();
        this.i.zp.top = +v;
        this.i.zp = this.i.zp;
    }
    ensureDataToolTipSummaryTitleTextMargin() {
        if (this.i.zp) {
            return;
        }
        this.i.zp = new Thickness(2);
    }
    get dataToolTipSummaryRowMarginBottom() {
        return this.i.zo ? this.i.zo.bottom : NaN;
    }
    set dataToolTipSummaryRowMarginBottom(v) {
        this.ensureDataToolTipSummaryRowMargin();
        this.i.zo.bottom = +v;
        this.i.zo = this.i.zo;
    }
    get dataToolTipSummaryRowMarginLeft() {
        return this.i.zo ? this.i.zo.left : NaN;
    }
    set dataToolTipSummaryRowMarginLeft(v) {
        this.ensureDataToolTipSummaryRowMargin();
        this.i.zo.left = +v;
        this.i.zo = this.i.zo;
    }
    get dataToolTipSummaryRowMarginRight() {
        return this.i.zo ? this.i.zo.right : NaN;
    }
    set dataToolTipSummaryRowMarginRight(v) {
        this.ensureDataToolTipSummaryRowMargin();
        this.i.zo.right = +v;
        this.i.zo = this.i.zo;
    }
    get dataToolTipSummaryRowMarginTop() {
        return this.i.zo ? this.i.zo.top : NaN;
    }
    set dataToolTipSummaryRowMarginTop(v) {
        this.ensureDataToolTipSummaryRowMargin();
        this.i.zo.top = +v;
        this.i.zo = this.i.zo;
    }
    ensureDataToolTipSummaryRowMargin() {
        if (this.i.zo) {
            return;
        }
        this.i.zo = new Thickness(2);
    }
    /**
     * Gets or sets the units text color.
    */
    get dataToolTipSummaryValueTextColor() {
        return brushToString(this.i.yj);
    }
    set dataToolTipSummaryValueTextColor(v) {
        this.i.yj = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get dataToolTipSummaryValueTextStyle() {
        if (this.i.fb == null) {
            return null;
        }
        return this.i.fb.fontString;
    }
    set dataToolTipSummaryValueTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.fb = fi;
    }
    /**
     * Gets or sets the units text for the data legend.
    */
    get dataToolTipSummaryLabelText() {
        return this.i.o7;
    }
    set dataToolTipSummaryLabelText(v) {
        this.i.o7 = v;
    }
    /**
     * Gets or sets the units text color.
    */
    get dataToolTipSummaryLabelTextColor() {
        return brushToString(this.i.yg);
    }
    set dataToolTipSummaryLabelTextColor(v) {
        this.i.yg = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get dataToolTipSummaryLabelTextStyle() {
        if (this.i.e8 == null) {
            return null;
        }
        return this.i.e8.fontString;
    }
    set dataToolTipSummaryLabelTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.e8 = fi;
    }
    /**
     * Gets or sets the units text for the data legend.
    */
    get dataToolTipSummaryUnitsText() {
        return this.i.pi;
    }
    set dataToolTipSummaryUnitsText(v) {
        this.i.pi = v;
    }
    /**
     * Gets or sets the units text color.
    */
    get dataToolTipSummaryUnitsTextColor() {
        return brushToString(this.i.yi);
    }
    set dataToolTipSummaryUnitsTextColor(v) {
        this.i.yi = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get dataToolTipSummaryUnitsTextStyle() {
        if (this.i.fa == null) {
            return null;
        }
        return this.i.fa.fontString;
    }
    set dataToolTipSummaryUnitsTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.fa = fi;
    }
    get dataToolTipBadgeMarginBottom() {
        return this.i.zi ? this.i.zi.bottom : NaN;
    }
    set dataToolTipBadgeMarginBottom(v) {
        this.ensureDataToolTipBadgeMargin();
        this.i.zi.bottom = +v;
        this.i.zi = this.i.zi;
    }
    get dataToolTipBadgeMarginLeft() {
        return this.i.zi ? this.i.zi.left : NaN;
    }
    set dataToolTipBadgeMarginLeft(v) {
        this.ensureDataToolTipBadgeMargin();
        this.i.zi.left = +v;
        this.i.zi = this.i.zi;
    }
    get dataToolTipBadgeMarginRight() {
        return this.i.zi ? this.i.zi.right : NaN;
    }
    set dataToolTipBadgeMarginRight(v) {
        this.ensureDataToolTipBadgeMargin();
        this.i.zi.right = +v;
        this.i.zi = this.i.zi;
    }
    get dataToolTipBadgeMarginTop() {
        return this.i.zi ? this.i.zi.top : NaN;
    }
    set dataToolTipBadgeMarginTop(v) {
        this.ensureDataToolTipBadgeMargin();
        this.i.zi.top = +v;
        this.i.zi = this.i.zi;
    }
    ensureDataToolTipBadgeMargin() {
        if (this.i.zi) {
            return;
        }
        this.i.zi = new Thickness(2);
    }
    /**
     * Gets or sets the BadgeShape for the data legend.
    */
    get dataToolTipBadgeShape() {
        return this.i.cs;
    }
    set dataToolTipBadgeShape(v) {
        this.i.cs = ensureEnum(LegendItemBadgeShape_$type, v);
    }
    /**
     * Gets or sets the UnitsMode for the data legend.
    */
    get dataToolTipUnitsDisplayMode() {
        return this.i.br;
    }
    set dataToolTipUnitsDisplayMode(v) {
        this.i.br = ensureEnum(DataLegendUnitsMode_$type, v);
    }
    /**
     * Gets or sets the units text for the data legend.
    */
    get dataToolTipUnitsText() {
        return this.i.ps;
    }
    set dataToolTipUnitsText(v) {
        this.i.ps = v;
    }
    /**
     * Gets or sets the units text color.
    */
    get dataToolTipUnitsTextColor() {
        return brushToString(this.i.yl);
    }
    set dataToolTipUnitsTextColor(v) {
        this.i.yl = stringToBrush(v);
    }
    get dataToolTipUnitsTextMarginBottom() {
        return this.i.zr ? this.i.zr.bottom : NaN;
    }
    set dataToolTipUnitsTextMarginBottom(v) {
        this.ensureDataToolTipUnitsTextMargin();
        this.i.zr.bottom = +v;
        this.i.zr = this.i.zr;
    }
    get dataToolTipUnitsTextMarginLeft() {
        return this.i.zr ? this.i.zr.left : NaN;
    }
    set dataToolTipUnitsTextMarginLeft(v) {
        this.ensureDataToolTipUnitsTextMargin();
        this.i.zr.left = +v;
        this.i.zr = this.i.zr;
    }
    get dataToolTipUnitsTextMarginRight() {
        return this.i.zr ? this.i.zr.right : NaN;
    }
    set dataToolTipUnitsTextMarginRight(v) {
        this.ensureDataToolTipUnitsTextMargin();
        this.i.zr.right = +v;
        this.i.zr = this.i.zr;
    }
    get dataToolTipUnitsTextMarginTop() {
        return this.i.zr ? this.i.zr.top : NaN;
    }
    set dataToolTipUnitsTextMarginTop(v) {
        this.ensureDataToolTipUnitsTextMargin();
        this.i.zr.top = +v;
        this.i.zr = this.i.zr;
    }
    ensureDataToolTipUnitsTextMargin() {
        if (this.i.zr) {
            return;
        }
        this.i.zr = new Thickness(2);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get dataToolTipUnitsTextStyle() {
        if (this.i.fd == null) {
            return null;
        }
        return this.i.fd.fontString;
    }
    set dataToolTipUnitsTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.fd = fi;
    }
    get dataToolTipTitleTextMarginBottom() {
        return this.i.zq ? this.i.zq.bottom : NaN;
    }
    set dataToolTipTitleTextMarginBottom(v) {
        this.ensureDataToolTipTitleTextMargin();
        this.i.zq.bottom = +v;
        this.i.zq = this.i.zq;
    }
    get dataToolTipTitleTextMarginLeft() {
        return this.i.zq ? this.i.zq.left : NaN;
    }
    set dataToolTipTitleTextMarginLeft(v) {
        this.ensureDataToolTipTitleTextMargin();
        this.i.zq.left = +v;
        this.i.zq = this.i.zq;
    }
    get dataToolTipTitleTextMarginRight() {
        return this.i.zq ? this.i.zq.right : NaN;
    }
    set dataToolTipTitleTextMarginRight(v) {
        this.ensureDataToolTipTitleTextMargin();
        this.i.zq.right = +v;
        this.i.zq = this.i.zq;
    }
    get dataToolTipTitleTextMarginTop() {
        return this.i.zq ? this.i.zq.top : NaN;
    }
    set dataToolTipTitleTextMarginTop(v) {
        this.ensureDataToolTipTitleTextMargin();
        this.i.zq.top = +v;
        this.i.zq = this.i.zq;
    }
    ensureDataToolTipTitleTextMargin() {
        if (this.i.zq) {
            return;
        }
        this.i.zq = new Thickness(2);
    }
    /**
     * Gets or sets the display text color.
    */
    get dataToolTipTitleTextColor() {
        return brushToString(this.i.yk);
    }
    set dataToolTipTitleTextColor(v) {
        this.i.yk = stringToBrush(v);
    }
    /**
     * Gets or Sets the style to use for the display text.
    */
    get dataToolTipTitleTextStyle() {
        if (this.i.fc == null) {
            return null;
        }
        return this.i.fc.fontString;
    }
    set dataToolTipTitleTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.fc = fi;
    }
    /**
     * Gets or sets the mode for displaying labels before series values in the data legend, e.g. O: H: L: C: for financial series
    */
    get dataToolTipLabelDisplayMode() {
        return this.i.bn;
    }
    set dataToolTipLabelDisplayMode(v) {
        this.i.bn = ensureEnum(DataLegendLabelMode_$type, v);
    }
    /**
     * Gets or sets the units text color.
    */
    get dataToolTipLabelTextColor() {
        return brushToString(this.i.yf);
    }
    set dataToolTipLabelTextColor(v) {
        this.i.yf = stringToBrush(v);
    }
    get dataToolTipLabelTextMarginBottom() {
        return this.i.zn ? this.i.zn.bottom : NaN;
    }
    set dataToolTipLabelTextMarginBottom(v) {
        this.ensureDataToolTipLabelTextMargin();
        this.i.zn.bottom = +v;
        this.i.zn = this.i.zn;
    }
    get dataToolTipLabelTextMarginLeft() {
        return this.i.zn ? this.i.zn.left : NaN;
    }
    set dataToolTipLabelTextMarginLeft(v) {
        this.ensureDataToolTipLabelTextMargin();
        this.i.zn.left = +v;
        this.i.zn = this.i.zn;
    }
    get dataToolTipLabelTextMarginRight() {
        return this.i.zn ? this.i.zn.right : NaN;
    }
    set dataToolTipLabelTextMarginRight(v) {
        this.ensureDataToolTipLabelTextMargin();
        this.i.zn.right = +v;
        this.i.zn = this.i.zn;
    }
    get dataToolTipLabelTextMarginTop() {
        return this.i.zn ? this.i.zn.top : NaN;
    }
    set dataToolTipLabelTextMarginTop(v) {
        this.ensureDataToolTipLabelTextMargin();
        this.i.zn.top = +v;
        this.i.zn = this.i.zn;
    }
    ensureDataToolTipLabelTextMargin() {
        if (this.i.zn) {
            return;
        }
        this.i.zn = new Thickness(2);
    }
    /**
     * Gets or Sets the style to use for the units text.
    */
    get dataToolTipLabelTextStyle() {
        if (this.i.e7 == null) {
            return null;
        }
        return this.i.e7.fontString;
    }
    set dataToolTipLabelTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.e7 = fi;
    }
    /**
     * Gets the domain type of this chart
    */
    get domainType() {
        return this.i.cb;
    }
    /**
     * Gets the current sort that is applied to the grid. Collection can be updated to modify the sort for the grid. Once sort descriptions are in this collection, the grid will no longer listen for changes on their properties.
    */
    get sortDescriptions() {
        const r = this.i.a3;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrChartSortDescriptionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    get groupSortDescriptions() {
        const r = this.i.az;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrChartSortDescriptionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets the current grouping that is applied to the grid. Collection can be updated to modify the grouping for the grid. Once grouping descriptions are in this collection, the grid will no longer listen for changes on their properties.
    */
    get groupDescriptions() {
        const r = this.i.ay;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrChartSortDescriptionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets the current filter that is applied to the chart. Collection can be updated to modify the filter for the chart. Once filter expressions are in this collection, the chart will no longer listen for changes on their properties.
    */
    get filterExpressions() {
        const r = this.i.ek;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrFilterExpressionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets the current highlight filter that is applied to the chart. Collection can be updated to modify the highlight filter for the chart. Once filter expressions are in this collection, the chart will no longer listen for changes on their properties.
    */
    get highlightFilterExpressions() {
        const r = this.i.el;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrFilterExpressionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets the current summaries that are applied to the grid.
    */
    get summaryDescriptions() {
        const r = this.i.a7;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrChartSummaryDescriptionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets or sets the selection mode to use for the series in the component, when supported.
    */
    get selectionMode() {
        return this.i.dx;
    }
    set selectionMode(v) {
        this.i.dx = ensureEnum(SeriesSelectionMode_$type, v);
    }
    /**
     * Gets or sets the focus mode to use for the series in the component, when supported.
    */
    get focusMode() {
        return this.i.dw;
    }
    set focusMode(v) {
        this.i.dw = ensureEnum(SeriesSelectionMode_$type, v);
    }
    /**
     * Gets or sets the selection brush to use for the series.
    */
    get selectionBrush() {
        return brushToString(this.i.yx);
    }
    set selectionBrush(v) {
        this.i.yx = stringToBrush(v);
    }
    /**
     * Gets or sets the focus brush to use for the series.
    */
    get focusBrush() {
        return brushToString(this.i.yp);
    }
    set focusBrush(v) {
        this.i.yp = stringToBrush(v);
    }
    /**
     * Gets or sets the selection behavior to use for the series in the component, when supported.
    */
    get selectionBehavior() {
        return this.i.dt;
    }
    set selectionBehavior(v) {
        this.i.dt = ensureEnum(SeriesSelectionBehavior_$type, v);
    }
    /**
     * Gets the currently selected data items.  Adding or removing data items from this collection will
     * select or deselect the visuals associated with those items.
    */
    get selectedSeriesItems() {
        if (this._selectedSeriesItems === null) {
            let coll = new IgrChartSelectedItemCollection();
            let innerColl = this.i.selectedSeriesItems;
            if (!innerColl) {
                innerColl = new ChartSelectedItemCollection_internal();
            }
            this._selectedSeriesItems = coll._fromInner(innerColl);
        }
        return this._selectedSeriesItems;
    }
    set selectedSeriesItems(v) {
        if (this._selectedSeriesItems !== null) {
            this._selectedSeriesItems._setSyncTarget(null);
            this._selectedSeriesItems = null;
        }
        let coll = new IgrChartSelectedItemCollection();
        this._selectedSeriesItems = coll._fromOuter(v);
        let syncColl = new SyncableObservableCollection$1(ChartSelection.$type);
        let innerColl = this.i.selectedSeriesItems;
        if (!innerColl) {
            innerColl = new ChartSelectedItemCollection_internal();
        }
        syncColl._inner = innerColl;
        syncColl.clear();
        this._selectedSeriesItems._setSyncTarget(syncColl);
    }
    /**
     * Gets the currently focused data items.  Adding or removing data items from this collection will
     * focus or blur the visuals associated with those items.
    */
    get focusedSeriesItems() {
        if (this._focusedSeriesItems === null) {
            let coll = new IgrChartSelectedItemCollection();
            let innerColl = this.i.focusedSeriesItems;
            if (!innerColl) {
                innerColl = new ChartSelectedItemCollection_internal();
            }
            this._focusedSeriesItems = coll._fromInner(innerColl);
        }
        return this._focusedSeriesItems;
    }
    set focusedSeriesItems(v) {
        if (this._focusedSeriesItems !== null) {
            this._focusedSeriesItems._setSyncTarget(null);
            this._focusedSeriesItems = null;
        }
        let coll = new IgrChartSelectedItemCollection();
        this._focusedSeriesItems = coll._fromOuter(v);
        let syncColl = new SyncableObservableCollection$1(ChartSelection.$type);
        let innerColl = this.i.focusedSeriesItems;
        if (!innerColl) {
            innerColl = new ChartSelectedItemCollection_internal();
        }
        syncColl._inner = innerColl;
        syncColl.clear();
        this._focusedSeriesItems._setSyncTarget(syncColl);
    }
    /**
     * Gets the initial sorts that are applied to the chart.
    */
    get initialSortDescriptions() {
        const r = this.i.a2;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrChartSortDescriptionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets the initial sorts that are applied to the chart after groupings and summaries are applied.
    */
    get initialGroupSortDescriptions() {
        const r = this.i.a1;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrChartSortDescriptionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets the initial groupings that are applied to the chart.
    */
    get initialGroupDescriptions() {
        const r = this.i.a0;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrChartSortDescriptionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets the initial filters that are applied to the chart.
    */
    get initialFilterExpressions() {
        const r = this.i.em;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrFilterExpressionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets the initial highlight filters that are applied to the chart.
    */
    get initialHighlightFilterExpressions() {
        const r = this.i.en;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrFilterExpressionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets the initial summaries that are applied to the chart.
    */
    get initialSummaryDescriptions() {
        const r = this.i.a6;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrChartSummaryDescriptionCollection();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    /**
     * Gets or sets the sorts to apply to the chart. This property will become ignored if sorts are changed outside of this property.
    */
    get initialSorts() {
        return this.i.rh;
    }
    set initialSorts(v) {
        this.i.rh = v;
    }
    /**
     * Gets or sets the sorts to apply after grouping has been applied.
    */
    get groupSorts() {
        return this.i.qq;
    }
    set groupSorts(v) {
        this.i.qq = v;
    }
    /**
     * Gets or sets the groupings to apply to the chart. This property will become ignored if sorts are changed outside of this property.
    */
    get initialGroups() {
        return this.i.rd;
    }
    set initialGroups(v) {
        this.i.rd = v;
    }
    /**
     * Gets or sets the filter to apply to the chart. This property will become ignored if the filter is changed outside of this property.
    */
    get initialFilter() {
        return this.i.rb;
    }
    set initialFilter(v) {
        this.i.rb = v;
    }
    /**
     * Gets or sets the filter to apply to the chart. This property will become ignored if the filter is changed outside of this property.
    */
    get initialHighlightFilter() {
        return this.i.rf;
    }
    set initialHighlightFilter(v) {
        this.i.rf = v;
    }
    /**
     * Gets or sets the summaries to apply to the chart. This property will become ignored if sorts are changed outside of this property.
    */
    get initialSummaries() {
        return this.i.rj;
    }
    set initialSummaries(v) {
        this.i.rj = v;
    }
    /**
     * Gets or sets a collection of data items used to generate the chart.
     * The HighlightedItemsSource of this chart can be a list of objects containing one or more numeric properties.
     * Additionally, if the objects in the list implement the IEnumerable interface,
     * the Chart will attempt to delve into the sub-collections when reading through the data source.
     * Data binding can be further configured by attributing the data item classes
     * with the DataSeriesMemberIntentAttribute.
    */
    get highlightedDataSource() {
        return this.i.highlightedItemsSource;
    }
    set highlightedDataSource(v) {
        this.i.highlightedItemsSource = v;
    }
    /**
     * Gets or sets a set of property paths that should be included for consideration by the category chart, leaving the remainder excluded. If null, all properties will be considered.
    *
    * The `IncludedProperties` property used to include the properties for the consideration of the category chart.
    *
    * ```ts
    * this.chart.includedProperties = ["ProductName", "Cost"];
    * ```
    *
    * ```ts
    * 	<IgrCategoryChart
    * 		chartType="Line"
    * 		markerTypes={this.state.markersTypes}
    * 		dataSource={this.categoryData}
    * 		excludedProperties="value" />
    * ```
    */
    get includedProperties() {
        return this.i.includedProperties;
    }
    set includedProperties(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.includedProperties = v;
    }
    /**
     * Gets or sets a set of property paths that should be excluded from consideration by the category chart.
    *
    * The `ExcludedProperties` property used for the property paths that should be excluded from consideration by the category chart.
    *
    * ```ts
    * this.chart.excludedProperties = ["ID", "Discount"];
    * ```
    *
    * ```ts
    * 	<IgrCategoryChart
    * 	chartType="Line"
    * 	markerTypes={this.state.markersTypes}
    * 	dataSource={this.categoryData}
    * 	excludedProperties="value" />
    * ```
    */
    get excludedProperties() {
        return this.i.excludedProperties;
    }
    set excludedProperties(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.excludedProperties = v;
    }
    /**
     * Gets or sets the palette of brushes to use for coloring the chart series.
     * The value provided should be an array of CSS color strings or JavaScript objects defining gradients. Optionally the first element can be a string reading "RGB" or "HSV" to specify the interpolation mode of the collection
    *
    * Use the `Brushes` property to  set the brushes.
    *
    * ```ts
    * this.chart.brushes = ["#ff0000","#00ff00"];
    * ```
    */
    get brushes() {
        return fromBrushCollection(this.i.z);
    }
    set brushes(v) {
        this.i.z = toBrushCollection(v);
    }
    /**
     * Gets or sets the palette of brushes to use for outlines on the chart series.
     * The value provided should be an array of CSS color strings or JavaScript objects defining gradients. Optionally the first element can be a string reading "RGB" or "HSV" to specify the interpolation mode of the collection
    *
    *  Use the `Outlines` property to sets the palette of brushes to use for outlines on the chart series
    *
    * ```ts
    * <IgrCategoryChart
    * 	chartType="Line"
    * 	outlines= "#ff0000"
    * 	dataSource={this.categoryData}/>
    * ```
    */
    get outlines() {
        return fromBrushCollection(this.i.ad);
    }
    set outlines(v) {
        this.i.ad = toBrushCollection(v);
    }
    /**
     * Gets the actual palette of brushes to use for coloring the chart series.
    */
    get actualBrushes() {
        return fromBrushCollection(this.i.x);
    }
    set actualBrushes(v) {
        this.i.x = toBrushCollection(v);
    }
    /**
     * Gets the actual palette of brushes to use for outlines on the chart series.
    */
    get actualOutlines() {
        return fromBrushCollection(this.i.y);
    }
    set actualOutlines(v) {
        this.i.y = toBrushCollection(v);
    }
    /**
     * Gets or sets whether the chart can be horizontally zoomed through user interactions.
    *
    * Use the `IsHorizontalZoomEnabled` property to allow chart horizontally zoomed .
    *
    * ```ts
    *  <IgrCategoryChart dataSource={this.state.data}
    *                    width="700px"
    *                    height="500px"
    *                    xAxisLabelTextStyle="10pt Verdana"
    *                    xAxisLabelTopMargin={5}
    *                    xAxisLabelTextColor="gray"
    *                    yAxisLabelTextStyle="10pt Verdana"
    *                    yAxisLabelRightMargin={5}
    *                    yAxisLabelTextColor="gray"
    * 				   isHorizontalZoomEnabled= true/>
    * ```
    *
    * ```ts
    *  <IgrCategoryChart dataSource={this.state.data}
    *                    IsHorizontalZoomEnabled= true />
    * ```
    *
    * ```ts
    * this.chart.isHorizontalZoomEnabled = true;
    * ```
    */
    get isHorizontalZoomEnabled() {
        return this.i.g3;
    }
    set isHorizontalZoomEnabled(v) {
        this.i.g3 = ensureBool(v);
    }
    /**
     * Gets or sets whether the chart can be vertically zoomed through user interactions.
    *
    * Use the `IsVerticalZoomEnabled` property to allow chart zoom vertically.
    *
    * ```ts
    *  this.chart.isVerticalZoomEnabled = true;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart dataSource={this.state.data}
    *                   isVerticalZoomEnabled= "true"/>
    * ```
    */
    get isVerticalZoomEnabled() {
        return this.i.hb;
    }
    set isVerticalZoomEnabled(v) {
        this.i.hb = ensureBool(v);
    }
    /**
     * Gets or sets whether the chart should display mock data when the data source has not been provided, if supported for this type of chart.
    */
    get shouldDisplayMockData() {
        return this.i.hf;
    }
    set shouldDisplayMockData(v) {
        this.i.hf = ensureBool(v);
    }
    /**
     * Gets or sets whether the chart should use a faded skeleton style for mock data.
    */
    get shouldUseSkeletonStyleForMockData() {
        return this.i.hi;
    }
    set shouldUseSkeletonStyleForMockData(v) {
        this.i.hi = ensureBool(v);
    }
    /**
     * Gets or sets whether and how to display highlighted values for the series. Note, this is distinct from the highlighting feature that indicates what is closest or under the mouse.
     * This property applies to Category Chart and Financial Chart controls.
    */
    get highlightedValuesDisplayMode() {
        return this.i.dg;
    }
    set highlightedValuesDisplayMode(v) {
        this.i.dg = ensureEnum(SeriesHighlightedValuesDisplayMode_$type, v);
    }
    /**
     * Gets or sets whether the chart can highlight series through user interactions.
     * This property applies to Category Chart and Financial Chart controls.
    */
    get isSeriesHighlightingEnabled() {
        return this.i.g6;
    }
    set isSeriesHighlightingEnabled(v) {
        this.i.g6 = ensureBool(v);
    }
    /**
     * Gets or sets whether the chart can highlight series through user interactions.
     * This property applies to Category Chart and Financial Chart controls.
    */
    get highlightedLegendItemVisibility() {
        return this.i.zw;
    }
    set highlightedLegendItemVisibility(v) {
        this.i.zw = ensureEnum(Visibility_$type, v);
    }
    /**
     * Gets or sets whether to show the legend items for the series.
    */
    get legendItemVisibility() {
        return this.i.zx;
    }
    set legendItemVisibility(v) {
        this.i.zx = ensureEnum(Visibility_$type, v);
    }
    /**
     * Gets or sets the rectangle representing the current scroll and zoom state of the chart.
     * WindowRect is expressed as a Rectangle with coordinates and sizes between 0 and 1.
    *
    * Use `WindowRect` property representing the current scroll and zoom state of the chart.
    *
    * ```ts
    * this.chart.windowRect =[0,0,1,1];
    * ```
    *
    * ```ts
    * <IgrCategoryChart dataSource={this.state.data}
    *                WindowRect={1,0,1,1} />
    * ```
    */
    get windowRect() {
        return fromRect(this.i.y5);
    }
    set windowRect(v) {
        this.i.y5 = toRect(v);
    }
    /**
     * Gets a rectangle representing the bounds of the plot area.
    *
    * `viewport` property representing the bounds of the plot area.
    *
    * ```ts
    * let viewport:Rect =  this.chart.viewport;
    * ```
    */
    get viewport() {
        return fromRect(this.i.y4);
    }
    /**
     * Gets or sets text to display above the plot area.
    *
    * Use the `Title` property to display the text above the plot area.
    *
    * ```ts
    *  this.chart.title= "This is a Title";
    * ```
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	title= "CategoryChart Subtitle"	/>
    * ```
    */
    get chartTitle() {
        return this.i.sz;
    }
    set chartTitle(v) {
        this.i.sz = v;
    }
    /**
     * Gets or sets text to display below the Title, above the plot area.
    *
    * Use the `Subtitle` property to display the text below  and above the plot area.
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	subtitle= "CategoryChart Subtitle"/>
    * ```
    *
    * ```ts
    *  this.chart.subtitle ="CategoryChart Subtitle" ;
    * ```
    */
    get subtitle() {
        return this.i.so;
    }
    set subtitle(v) {
        this.i.so = v;
    }
    /**
     * Gets or sets horizontal alignment which determines the title position, relative to the left and right edges of the control.
    *
    * Use the `TitleAlignment` property for the horizontal alignment of the title.
    *
    * ```ts
    * this.chart.titleAlignment = HorizontalAlignment.Center;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	titleAlignment= "Center"/>
    * ```
    */
    get titleAlignment() {
        return this.i.w4;
    }
    set titleAlignment(v) {
        this.i.w4 = ensureEnum(HorizontalAlignment_$type, v);
    }
    /**
     * Gets or sets horizontal alignment which determines the subtitle position, relative to the left and right edges of the control.
    *
    * Use the `SubtitleAlignment` property to Gets or sets horizontal alignment.
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	dataSource={this.categoryData}
    * 	subtitle= "CategoryChart Subtitle"
    * 	subtitleAlignment="Right"/>
    * ```
    *
    * ```ts
    * this.chart.subtitleAlignment = HorizontalAlignment.Right;
    * ```
    */
    get subtitleAlignment() {
        return this.i.w3;
    }
    set subtitleAlignment(v) {
        this.i.w3 = ensureEnum(HorizontalAlignment_$type, v);
    }
    /**
     * Gets or sets the behavior that determines how unknown values will be plotted on the chart.
     * Null and Double.NaN are two examples of unknown values.
    *
    * Use the `UnknownValuePlotting` property to determines the behavior that how unknown values will be plotted on the chart.
    *
    * ```ts
    * this.chart.unknownValuePlotting =UnknownValuePlotting.DontPlot;
    * ```
    *
    * ```ts
    * <IgrCategoryChart dataSource={this.state.data}
    *                unknownValuePlotting= "dontPlot" />
    * ```
    */
    get unknownValuePlotting() {
        return this.i.ec;
    }
    set unknownValuePlotting(v) {
        this.i.ec = ensureEnum(UnknownValuePlotting_$type, v);
    }
    /**
     * Gets or sets the rendering resolution for all series in this chart.
     * Where n = Resolution, for every n horizontal pixels, combine all items into a single data point.  When Resolution = 0, all data points will be rendered as graphical objects.  Charts with a higher resolution will have faster performance.
    *
    * Use the `Resolution` property if the callouts should be display.
    *
    * ```ts
    *   this.chart.resolution = 2 ;
    * ```
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	markerTypes={this.state.markersTypes}
    * 	dataSource={this.categoryData}
    * 	calloutsVisible="true"
    * 	resolution= {2}/>
    * ```
    */
    get resolution() {
        return this.i.j5;
    }
    set resolution(v) {
        this.i.j5 = +v;
    }
    /**
     * Gets or sets the thickness for all series in this chart. Depending on the ChartType, this can be the main brush used, or just the outline.
    *
    * Use the `Thickness` property for the thickness of all the series in this chart.
    */
    get thickness() {
        return this.i.kb;
    }
    set thickness(v) {
        this.i.kb = +v;
    }
    /**
     * Gets or sets the OutlineMode for all series in this chart.
    */
    get outlineMode() {
        return this.i.dm;
    }
    set outlineMode(v) {
        this.i.dm = ensureEnum(SeriesOutlineMode_$type, v);
    }
    /**
     * Gets or sets the MarkerOutlineMode for all series that support markers in this chart.
    */
    get markerOutlineMode() {
        return this.i.cz;
    }
    set markerOutlineMode(v) {
        this.i.cz = ensureEnum(MarkerOutlineMode_$type, v);
    }
    /**
     * Gets or sets the MarkerFillMode for all series that support markers in this chart.
    */
    get markerFillMode() {
        return this.i.cx;
    }
    set markerFillMode(v) {
        this.i.cx = ensureEnum(MarkerFillMode_$type, v);
    }
    /**
     * Gets or sets the Marker Fill Opacity for all series that support markers in this chart.
    */
    get markerFillOpacity() {
        return this.i.jy;
    }
    set markerFillOpacity(v) {
        this.i.jy = +v;
    }
    /**
     * Gets or sets the Marker Thickness for all series that support markers in this chart.
    */
    get markerThickness() {
        return this.i.jz;
    }
    set markerThickness(v) {
        this.i.jz = +v;
    }
    /**
     * Gets or sets the marker shapes used for indicating location of data points in this chart.
     * This property applies only to these chart types: point, line, spline, bubble, and polygon
    */
    get markerTypes() {
        if (this._markerTypes === null) {
            let coll = new IgrMarkerTypeCollection();
            let innerColl = this.i.c5;
            if (!innerColl) {
                innerColl = new MarkerTypeCollection_internal();
            }
            this._markerTypes = coll._fromInner(innerColl);
            this.i.c5 = innerColl;
        }
        return this._markerTypes;
    }
    set markerTypes(v) {
        if (this._markerTypes !== null) {
            this._markerTypes._setSyncTarget(null);
            this._markerTypes = null;
        }
        let coll = new IgrMarkerTypeCollection();
        this._markerTypes = coll._fromOuter(v);
        let syncColl = new SyncableObservableCollection$1(MarkerType_$type);
        let innerColl = this.i.c5;
        if (!innerColl) {
            innerColl = new MarkerTypeCollection_internal();
        }
        syncColl._inner = innerColl;
        syncColl.clear();
        this._markerTypes._setSyncTarget(syncColl);
        this.i.c5 = innerColl;
    }
    /**
     * Gets or sets the default marker type for all series plotted in this chart.
     * This property is ignored when the MarkerTypes property is set
    */
    get markerAutomaticBehavior() {
        return this.i.cv;
    }
    set markerAutomaticBehavior(v) {
        this.i.cv = ensureEnum(MarkerAutomaticBehavior_$type, v);
    }
    /**
     * Gets or sets the palette of brushes used for rendering fill area of data point markers.
     * This property applies only to these chart types: point, line, spline, bubble, and polygon
    *
    * `MarkerBrushes` property used for rendering fill area of data point markers.
    *
    * ```ts
    *  this.chart.markerBrushes=[ "#ff0000", "#ffff00", "#00ffff" ];
    * ```
    *
    * ```ts
    * <IgrCategoryChart
    *     chartType="Line"
    *     markerBrushes ="#ffff00"
    *     dataSource={this.categoryData}/>
    * ```
    */
    get markerBrushes() {
        return fromBrushCollection(this.i.ab);
    }
    set markerBrushes(v) {
        this.i.ab = toBrushCollection(v);
    }
    /**
     * Gets or sets the palette of brushes used for rendering outlines of data point markers.
     * This property applies only to these chart types: point, line, spline, bubble, and polygon
    *
    * `MarkerOutlines` property used for rendering outlines of data point markers.
    *
    * ```ts
    * this.chart.markerOutlines= ["#ff0000", "#ffff00", "#00ffff"] ;
    * ```
    *
    * ```ts
    * <IgrCategoryChart
    * 	chartType="Line"
    * 	markerOutlines= "#ff0000"
    * 	dataSource={this.categoryData}/>
    * ```
    */
    get markerOutlines() {
        return fromBrushCollection(this.i.ac);
    }
    set markerOutlines(v) {
        this.i.ac = toBrushCollection(v);
    }
    /**
     * Gets or sets the maximum number of markers displayed in the plot area of the chart.
    *
    * `markerMaxCount` property used to display maximum number of markers in the plot area of the chart.
    *
    * ```ts
    *  this.chart.markerMaxCount = 100 ;
    * ```
    *
    * ```ts
    * <IgrCategoryChart
    * 	chartType="Line"
    * 	markerMaxCount= {2}
    * 	dataSource={this.categoryData}/>
    * ```
    */
    get markerMaxCount() {
        return this.i.lh;
    }
    set markerMaxCount(v) {
        this.i.lh = +v;
    }
    /**
     * Gets or sets the fill opacity for all series that have area visuals in this chart.
     * For example, Area, SplineArea, Column chart types
    */
    get areaFillOpacity() {
        return this.i.ji;
    }
    set areaFillOpacity(v) {
        this.i.ji = +v;
    }
    /**
     * Gets or sets whether the series animations should be allowed when a range change has been detected on an axis.
    */
    get animateSeriesWhenAxisRangeChanges() {
        return this.i.gg;
    }
    set animateSeriesWhenAxisRangeChanges(v) {
        this.i.gg = ensureBool(v);
    }
    /**
     * Gets or sets the palette of brushes to used for coloring trend lines in this chart.
     * The value provided should be an array of CSS color strings or JavaScript objects defining gradients. Optionally the first element can be a string reading "RGB" or "HSV" to specify the interpolation mode of the collection
    *
    * Use the `TrendLineBrushes` property for coloring trend lines in the chart.
    *
    * ```ts
    * this.chart.trendLineBrushes=["#ff0000", "#ffff00", "#00ffff"]	;
    * ```
    *
    * ```ts
    * <IgrCategoryChart dataSource={this.state.data}
    *                trendLineBrushes="#ffff00" />
    * ```
    */
    get trendLineBrushes() {
        return fromBrushCollection(this.i.af);
    }
    set trendLineBrushes(v) {
        this.i.af = toBrushCollection(v);
    }
    /**
     * Gets or sets the formula used for calculating trend lines in this chart.
     * This property applies only to these chart types: point, line, spline, and bubble
    *
    * Use the `TrendLineType` property for calculating trend lines.
    *
    * ```ts
    * this.chart.trendLineType = TrendLineType.CubicFit;
    * ```
    *
    * ```ts
    * <IgrCategoryChart dataSource={this.state.data}
    *                trendLineType="CubicFit" />
    * ```
    */
    get trendLineType() {
        return this.i.trendLineType;
    }
    set trendLineType(v) {
        this.i.trendLineType = ensureEnum(TrendLineType_$type, v);
    }
    /**
     * Gets or sets the thickness of the trend lines in this chart.
     * This property applies only to these chart types: point, line, spline, and bubble
    *
    * Use the `TrendLineThickness` property for the thickness of the trend lines in the chart.
    *
    * ```ts
    * this.chart.trendLineThickness=2;
    * ```
    *
    * ```ts
    * <IgrCategoryChart dataSource={this.state.data}
    *                trendLineThickness={2} />
    * ```
    */
    get trendLineThickness() {
        return this.i.kh;
    }
    set trendLineThickness(v) {
        this.i.kh = +v;
    }
    /**
     * Gets or sets a value indicating whether grid and tick lines are aligned to device pixels.
    *
    * Use `AlignsGridLinesToPixels` propert to indicat whether grid and tick lines are aligned to device pixels.
    *
    * ```ts
    *  this.chart.alignsGridLinesToPixels= true;
    * ```
    */
    get alignsGridLinesToPixels() {
        return this.i.gf;
    }
    set alignsGridLinesToPixels(v) {
        this.i.gf = ensureBool(v);
    }
    /**
     * Margin applied to the Left of the plot area.
    */
    get plotAreaMarginLeft() {
        return this.i.j2;
    }
    set plotAreaMarginLeft(v) {
        this.i.j2 = +v;
    }
    /**
     * Margin applied to the Top of the plot area.
    */
    get plotAreaMarginTop() {
        return this.i.j4;
    }
    set plotAreaMarginTop(v) {
        this.i.j4 = +v;
    }
    /**
     * Margin applied to the right of the plot area.
    */
    get plotAreaMarginRight() {
        return this.i.j3;
    }
    set plotAreaMarginRight(v) {
        this.i.j3 = +v;
    }
    /**
     * Margin applied below the plot area.
    */
    get plotAreaMarginBottom() {
        return this.i.j1;
    }
    set plotAreaMarginBottom(v) {
        this.i.j1 = +v;
    }
    /**
     * Gets or sets mode to use for automatically calculating the plot area margin.
    */
    get computedPlotAreaMarginMode() {
        return this.i.a9;
    }
    set computedPlotAreaMarginMode(v) {
        this.i.a9 = ensureEnum(ComputedPlotAreaMarginMode_$type, v);
    }
    /**
     * Gets or sets horizontal mode to use for automatically calculating the plot area margin based ont the series.
    */
    get seriesPlotAreaMarginHorizontalMode() {
        return this.i.dp;
    }
    set seriesPlotAreaMarginHorizontalMode(v) {
        this.i.dp = ensureEnum(SeriesPlotAreaMarginHorizontalMode_$type, v);
    }
    /**
     * Gets or sets vertical mode to use for automatically calculating the plot area margin based ont the series.
    */
    get seriesPlotAreaMarginVerticalMode() {
        return this.i.dr;
    }
    set seriesPlotAreaMarginVerticalMode(v) {
        this.i.dr = ensureEnum(SeriesPlotAreaMarginVerticalMode_$type, v);
    }
    /**
     * Gets or sets the highlighting mode to use for the series in the component, when supported. This takes precedence over the series level IsHighlightingEnabled.
    */
    get highlightingMode() {
        return this.i.dk;
    }
    set highlightingMode(v) {
        this.i.dk = ensureEnum(SeriesHighlightingMode_$type, v);
    }
    /**
     * Gets or sets the highlighting Behavior to use for the series in the component, when supported. This takes precedence over the series level IsHighlightingEnabled.
    */
    get highlightingBehavior() {
        return this.i.di;
    }
    set highlightingBehavior(v) {
        this.i.di = ensureEnum(SeriesHighlightingBehavior_$type, v);
    }
    /**
     * Gets or sets the highlighting mode to use for the legend linked to the component, when supported.
    */
    get legendHighlightingMode() {
        return this.i.cn;
    }
    set legendHighlightingMode(v) {
        this.i.cn = ensureEnum(LegendHighlightingMode_$type, v);
    }
    /**
     * Gets or sets the type of legend badges representing all series displayed in a legend linked to this component
     * This property will be ignored when the LegendItemTemplate or LegendItemBadgeTemplate property is set on a series
    */
    get legendItemBadgeShape() {
        return this.i.ct;
    }
    set legendItemBadgeShape(v) {
        this.i.ct = ensureEnum(LegendItemBadgeShape_$type, v);
    }
    /**
     * Gets or sets the mode of legend badges representing all series displayed in a legend linked to this component.
     * This property will be ignored when the LegendItemTemplate or LegendItemBadgeTemplate property is set on a series
    */
    get legendItemBadgeMode() {
        return this.i.cp;
    }
    set legendItemBadgeMode(v) {
        this.i.cp = ensureEnum(LegendItemBadgeMode_$type, v);
    }
    get trendLinePeriod() {
        return this.i.lk;
    }
    set trendLinePeriod(v) {
        this.i.lk = +v;
    }
    /**
     * Gets or sets the style of tooltip to be displayed.
    */
    get toolTipType() {
        return this.i.d9;
    }
    set toolTipType(v) {
        this.i.d9 = ensureEnum(ToolTipType_$type, v);
    }
    /**
     * Gets or sets the crosshairs to be displayed.
    */
    get crosshairsDisplayMode() {
        return this.i.crosshairsDisplayMode;
    }
    set crosshairsDisplayMode(v) {
        this.i.crosshairsDisplayMode = ensureEnum(CrosshairsDisplayMode_$type, v);
    }
    /**
     * Gets or sets whether crosshairs will snap to the nearest data point.
    */
    get crosshairsSnapToData() {
        return this.i.gn;
    }
    set crosshairsSnapToData(v) {
        this.i.gn = ensureBool(v);
    }
    /**
     * Gets or sets the color to apply to vertical crosshairs line.
    */
    get crosshairsLineVerticalStroke() {
        return brushToString(this.i.yc);
    }
    set crosshairsLineVerticalStroke(v) {
        this.i.yc = stringToBrush(v);
    }
    /**
     * Gets or sets thickness of crosshairs lines.
    */
    get crosshairsLineThickness() {
        return this.i.jl;
    }
    set crosshairsLineThickness(v) {
        this.i.jl = +v;
    }
    /**
     * Gets or sets the color to apply to horizontal crosshairs line.
    */
    get crosshairsLineHorizontalStroke() {
        return brushToString(this.i.yb);
    }
    set crosshairsLineHorizontalStroke(v) {
        this.i.yb = stringToBrush(v);
    }
    /**
     * Gets or sets whether annotations are shown along the axis for crosshair values
    */
    get crosshairsAnnotationEnabled() {
        return this.i.crosshairsAnnotationEnabled;
    }
    set crosshairsAnnotationEnabled(v) {
        this.i.crosshairsAnnotationEnabled = ensureBool(v);
    }
    /**
     * Gets or sets the background of crosshair annotation on x-axis.
    */
    get crosshairsAnnotationXAxisBackground() {
        return brushToString(this.i.x7);
    }
    set crosshairsAnnotationXAxisBackground(v) {
        this.i.x7 = stringToBrush(v);
    }
    /**
     * Gets or sets the background of crosshair annotation on y-axis.
    */
    get crosshairsAnnotationYAxisBackground() {
        return brushToString(this.i.x9);
    }
    set crosshairsAnnotationYAxisBackground(v) {
        this.i.x9 = stringToBrush(v);
    }
    /**
     * Gets or sets the text color of crosshair annotation on x-axis.
    */
    get crosshairsAnnotationXAxisTextColor() {
        return brushToString(this.i.x8);
    }
    set crosshairsAnnotationXAxisTextColor(v) {
        this.i.x8 = stringToBrush(v);
    }
    /**
     * Gets or sets the text color of crosshair annotation on y-axis.
    */
    get crosshairsAnnotationYAxisTextColor() {
        return brushToString(this.i.ya);
    }
    set crosshairsAnnotationYAxisTextColor(v) {
        this.i.ya = stringToBrush(v);
    }
    /**
     * Gets or sets precision on interpolated values of crosshairs on x-axis.
    */
    get crosshairsAnnotationXAxisPrecision() {
        return this.i.la;
    }
    set crosshairsAnnotationXAxisPrecision(v) {
        this.i.la = +v;
    }
    /**
     * Gets or sets precision on interpolated values of crosshairs on y-axis.
    */
    get crosshairsAnnotationYAxisPrecision() {
        return this.i.lb;
    }
    set crosshairsAnnotationYAxisPrecision(v) {
        this.i.lb = +v;
    }
    /**
     * Gets or sets whether annotations for the final value of each series is displayed on the axis.
    */
    get shouldAvoidAxisAnnotationCollisions() {
        return this.i.shouldAvoidAxisAnnotationCollisions;
    }
    set shouldAvoidAxisAnnotationCollisions(v) {
        this.i.shouldAvoidAxisAnnotationCollisions = ensureBool(v);
    }
    /**
     * Gets or sets a whether the chart should pan its content when zoo,ing in passed max zoom level.
    */
    get shouldPanOnMaximumZoom() {
        return this.i.hg;
    }
    set shouldPanOnMaximumZoom(v) {
        this.i.hg = ensureBool(v);
    }
    /**
     * Gets or sets whether annotations for the final value of each series is displayed on the axis.
    */
    get finalValueAnnotationsVisible() {
        return this.i.finalValueAnnotationsVisible;
    }
    set finalValueAnnotationsVisible(v) {
        this.i.finalValueAnnotationsVisible = ensureBool(v);
    }
    /**
     * Gets or sets the background of final value annotation.
    */
    get finalValueAnnotationsBackground() {
        return brushToString(this.i.yn);
    }
    set finalValueAnnotationsBackground(v) {
        this.i.yn = stringToBrush(v);
    }
    /**
     * Gets or sets the text color of final value annotation.
    */
    get finalValueAnnotationsTextColor() {
        return brushToString(this.i.yo);
    }
    set finalValueAnnotationsTextColor(v) {
        this.i.yo = stringToBrush(v);
    }
    /**
     * Gets or sets precision on final value annotation.
    */
    get finalValueAnnotationsPrecision() {
        return this.i.le;
    }
    set finalValueAnnotationsPrecision(v) {
        this.i.le = +v;
    }
    /**
     * Gets or sets if the auto value callouts should be displayed.
    */
    get autoCalloutsVisible() {
        return this.i.autoCalloutsVisible;
    }
    set autoCalloutsVisible(v) {
        this.i.autoCalloutsVisible = ensureBool(v);
    }
    /**
     * Gets or sets if callouts should be displayed.
    */
    get calloutsVisible() {
        return this.i.gm;
    }
    set calloutsVisible(v) {
        this.i.gm = ensureBool(v);
    }
    /**
     * Gets or sets if event annotations should be displayed.
    */
    get calloutStyleUpdatingEventEnabled() {
        return this.i.gi;
    }
    set calloutStyleUpdatingEventEnabled(v) {
        this.i.gi = ensureBool(v);
    }
    /**
     * Gets or sets if event annotations should be displayed.
    */
    get useValueForAutoCalloutLabels() {
        return this.i.hk;
    }
    set useValueForAutoCalloutLabels(v) {
        this.i.hk = ensureBool(v);
    }
    /**
     * Gets or sets if event annotations should be displayed.
    */
    get calloutCollisionMode() {
        return this.i.ah;
    }
    set calloutCollisionMode(v) {
        this.i.ah = ensureEnum(CalloutCollisionMode_$type, v);
    }
    /**
     * Gets or sets if the series item color should be used for the callout fills.
    */
    get calloutsUseItemColorForFill() {
        return this.i.gk;
    }
    set calloutsUseItemColorForFill(v) {
        this.i.gk = ensureBool(v);
    }
    /**
     * Gets or sets if the series item color should be used for the callout outlines.
    */
    get calloutsUseItemColorForOutline() {
        return this.i.gl;
    }
    set calloutsUseItemColorForOutline(v) {
        this.i.gl = ensureBool(v);
    }
    /**
     * Gets or sets whether to the label color is automatically decided between a light and dark option for contrast.
    */
    get calloutsUseAutoContrastingLabelColors() {
        return this.i.gj;
    }
    set calloutsUseAutoContrastingLabelColors(v) {
        this.i.gj = ensureBool(v);
    }
    /**
     * Gets or sets if the thickness that should be used for the callout outlines.
    */
    get calloutsStrokeThickness() {
        return this.i.jk;
    }
    set calloutsStrokeThickness(v) {
        this.i.jk = +v;
    }
    /**
     * Gets or sets color of the callout background.
    */
    get calloutsBackground() {
        return brushToString(this.i.x1);
    }
    set calloutsBackground(v) {
        this.i.x1 = stringToBrush(v);
    }
    /**
     * Gets or sets color of the callout outlines.
    */
    get calloutsOutline() {
        return brushToString(this.i.x5);
    }
    set calloutsOutline(v) {
        this.i.x5 = stringToBrush(v);
    }
    /**
     * Gets or sets text color of callout.
    */
    get calloutsTextColor() {
        return brushToString(this.i.x6);
    }
    set calloutsTextColor(v) {
        this.i.x6 = stringToBrush(v);
    }
    /**
     * Gets or sets the text color on dark callout background.
    */
    get calloutsLightTextColor() {
        return brushToString(this.i.x4);
    }
    set calloutsLightTextColor(v) {
        this.i.x4 = stringToBrush(v);
    }
    /**
     * Gets or sets the text color on light callout background.
    */
    get calloutsDarkTextColor() {
        return brushToString(this.i.x2);
    }
    set calloutsDarkTextColor(v) {
        this.i.x2 = stringToBrush(v);
    }
    /**
     * Gets or sets color of callout's leader lines.
    */
    get calloutsLeaderBrush() {
        return brushToString(this.i.x3);
    }
    set calloutsLeaderBrush(v) {
        this.i.x3 = stringToBrush(v);
    }
    /**
     * Gets or sets the amount of precision to apply to the auto callout labels.
    */
    get calloutsAutoLabelPrecision() {
        return this.i.k9;
    }
    set calloutsAutoLabelPrecision(v) {
        this.i.k9 = +v;
    }
    /**
     * Gets or sets the allowed positions that the callout layer is allowed to use.
    */
    get calloutsAllowedPositions() {
        if (this._calloutsAllowedPositions === null) {
            let coll = new IgrCalloutPlacementPositionsCollection();
            let innerColl = this.i.aj;
            if (!innerColl) {
                innerColl = new CalloutPlacementPositionsCollection_internal();
            }
            this._calloutsAllowedPositions = coll._fromInner(innerColl);
            this.i.aj = innerColl;
        }
        return this._calloutsAllowedPositions;
    }
    set calloutsAllowedPositions(v) {
        if (this._calloutsAllowedPositions !== null) {
            this._calloutsAllowedPositions._setSyncTarget(null);
            this._calloutsAllowedPositions = null;
        }
        let coll = new IgrCalloutPlacementPositionsCollection();
        this._calloutsAllowedPositions = coll._fromOuter(v);
        let syncColl = new SyncableObservableCollection$1(CalloutPlacementPositions_$type);
        let innerColl = this.i.aj;
        if (!innerColl) {
            innerColl = new CalloutPlacementPositionsCollection_internal();
        }
        syncColl._inner = innerColl;
        syncColl.clear();
        this._calloutsAllowedPositions._setSyncTarget(syncColl);
        this.i.aj = innerColl;
    }
    /**
     * Gets or sets the member path of the X data for the callouts.
    */
    get calloutsXMemberPath() {
        return this.i.no;
    }
    set calloutsXMemberPath(v) {
        this.i.no = v;
    }
    /**
     * Gets or sets the member path of the Y data for the callouts.
    */
    get calloutsYMemberPath() {
        return this.i.nq;
    }
    set calloutsYMemberPath(v) {
        this.i.nq = v;
    }
    /**
     * Gets or sets the member path of the label data for the callouts.
    */
    get calloutsLabelMemberPath() {
        return this.i.nb;
    }
    set calloutsLabelMemberPath(v) {
        this.i.nb = v;
    }
    /**
     * Gets or sets the member path of the content data for the callouts.
    */
    get calloutsContentMemberPath() {
        return this.i.m7;
    }
    set calloutsContentMemberPath(v) {
        this.i.m7 = v;
    }
    /**
     * Gets or sets the font for callout text.
    */
    get calloutsTextStyle() {
        if (this.i.e4 == null) {
            return null;
        }
        return this.i.e4.fontString;
    }
    set calloutsTextStyle(v) {
        let fi = new FontInfo();
        fi.fontString = v;
        this.i.e4 = fi;
    }
    /**
     * Gets or sets what value lines to use in the chart.
    */
    get valueLines() {
        if (this._valueLines === null) {
            let coll = new IgrValueModeCollection();
            let innerColl = this.i.valueLines;
            if (!innerColl) {
                innerColl = new ValueModeCollection_internal();
            }
            this._valueLines = coll._fromInner(innerColl);
            this.i.valueLines = innerColl;
        }
        return this._valueLines;
    }
    set valueLines(v) {
        if (this._valueLines !== null) {
            this._valueLines._setSyncTarget(null);
            this._valueLines = null;
        }
        let coll = new IgrValueModeCollection();
        this._valueLines = coll._fromOuter(v);
        let syncColl = new SyncableObservableCollection$1(ValueLayerValueMode_$type);
        let innerColl = this.i.valueLines;
        if (!innerColl) {
            innerColl = new ValueModeCollection_internal();
        }
        syncColl._inner = innerColl;
        syncColl.clear();
        this._valueLines._setSyncTarget(syncColl);
        this.i.valueLines = innerColl;
    }
    /**
     * Gets or sets the palette of brushes to use for coloring the value lines present in the ValueLines property.
     * The value provided should be an array of CSS color strings or JavaScript objects defining gradients. Optionally the first element can be a string reading "RGB" or "HSV" to specify the interpolation mode of the collection
    */
    get valueLinesBrushes() {
        return fromBrushCollection(this.i.valueLinesBrushes);
    }
    set valueLinesBrushes(v) {
        this.i.valueLinesBrushes = toBrushCollection(v);
    }
    /**
     * Gets or sets thickness of value lines.
    */
    get valueLinesThickness() {
        return this.i.ki;
    }
    set valueLinesThickness(v) {
        this.i.ki = +v;
    }
    /**
     * Gets or sets if the chart should use the legend for value layers that target series.
    */
    get seriesValueLayerUseLegend() {
        return this.i.he;
    }
    set seriesValueLayerUseLegend(v) {
        this.i.he = ensureBool(v);
    }
    /**
     * Gets or sets the horizontal scrollbar mode to use for the series viewer.
    */
    get horizontalViewScrollbarMode() {
        return this.i.d2;
    }
    set horizontalViewScrollbarMode(v) {
        this.i.d2 = ensureEnum(SeriesViewerScrollbarMode_$type, v);
    }
    /**
     * Gets or sets the vertical scrollbar mode to use for the series viewer.
    */
    get verticalViewScrollbarMode() {
        return this.i.d3;
    }
    set verticalViewScrollbarMode(v) {
        this.i.d3 = ensureEnum(SeriesViewerScrollbarMode_$type, v);
    }
    /**
     * Gets or sets the position to use for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarPosition() {
        return this.i.dz;
    }
    set horizontalViewScrollbarPosition(v) {
        this.i.dz = ensureEnum(SeriesViewerHorizontalScrollbarPosition_$type, v);
    }
    /**
     * Gets or sets the position to use for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarPosition() {
        return this.i.d7;
    }
    set verticalViewScrollbarPosition(v) {
        this.i.d7 = ensureEnum(SeriesViewerVerticalScrollbarPosition_$type, v);
    }
    /**
     * Gets or sets the fill to use for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarFill() {
        return brushToString(this.i.yu);
    }
    set horizontalViewScrollbarFill(v) {
        this.i.yu = stringToBrush(v);
    }
    /**
     * Gets or sets the outline to use for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarOutline() {
        return brushToString(this.i.yv);
    }
    set horizontalViewScrollbarOutline(v) {
        this.i.yv = stringToBrush(v);
    }
    /**
     * Gets or sets the stroke thickness to use for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarStrokeThickness() {
        return this.i.ju;
    }
    set horizontalViewScrollbarStrokeThickness(v) {
        this.i.ju = +v;
    }
    /**
     * Gets or sets the max opacity to use for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarMaxOpacity() {
        return this.i.jt;
    }
    set horizontalViewScrollbarMaxOpacity(v) {
        this.i.jt = +v;
    }
    /**
     * Gets or sets the corner radius to use for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarCornerRadius() {
        return this.i.jq;
    }
    set horizontalViewScrollbarCornerRadius(v) {
        this.i.jq = +v;
    }
    /**
     * Gets or sets the height to use for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarHeight() {
        return this.i.jr;
    }
    set horizontalViewScrollbarHeight(v) {
        this.i.jr = +v;
    }
    /**
     * Gets or sets the inset distance to use for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarInset() {
        return this.i.js;
    }
    set horizontalViewScrollbarInset(v) {
        this.i.js = +v;
    }
    /**
     * Gets or sets the track start inset to use for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarTrackStartInset() {
        return this.i.jw;
    }
    set horizontalViewScrollbarTrackStartInset(v) {
        this.i.jw = +v;
    }
    /**
     * Gets or sets the track end inset to use for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarTrackEndInset() {
        return this.i.jv;
    }
    set horizontalViewScrollbarTrackEndInset(v) {
        this.i.jv = +v;
    }
    /**
     * Gets or sets whether to use automatic track insets for the horizontal scrollbar in the component, when enabled.
    */
    get horizontalViewScrollbarShouldAddAutoTrackInsets() {
        return this.i.gw;
    }
    set horizontalViewScrollbarShouldAddAutoTrackInsets(v) {
        this.i.gw = ensureBool(v);
    }
    /**
     * Gets or sets the fill to use for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarFill() {
        return brushToString(this.i.y0);
    }
    set verticalViewScrollbarFill(v) {
        this.i.y0 = stringToBrush(v);
    }
    /**
     * Gets or sets the outline to use for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarOutline() {
        return brushToString(this.i.y1);
    }
    set verticalViewScrollbarOutline(v) {
        this.i.y1 = stringToBrush(v);
    }
    /**
     * Gets or sets the stroke thickness to use for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarStrokeThickness() {
        return this.i.km;
    }
    set verticalViewScrollbarStrokeThickness(v) {
        this.i.km = +v;
    }
    /**
     * Gets or sets the max opacity to use for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarMaxOpacity() {
        return this.i.kl;
    }
    set verticalViewScrollbarMaxOpacity(v) {
        this.i.kl = +v;
    }
    /**
     * Gets or sets the corner radius to use for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarCornerRadius() {
        return this.i.kj;
    }
    set verticalViewScrollbarCornerRadius(v) {
        this.i.kj = +v;
    }
    /**
     * Gets or sets the width to use for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarWidth() {
        return this.i.kp;
    }
    set verticalViewScrollbarWidth(v) {
        this.i.kp = +v;
    }
    /**
     * Gets or sets the inset distance to use for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarInset() {
        return this.i.kk;
    }
    set verticalViewScrollbarInset(v) {
        this.i.kk = +v;
    }
    /**
     * Gets or sets the track start inset to use for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarTrackStartInset() {
        return this.i.ko;
    }
    set verticalViewScrollbarTrackStartInset(v) {
        this.i.ko = +v;
    }
    /**
     * Gets or sets the track end inset to use for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarTrackEndInset() {
        return this.i.kn;
    }
    set verticalViewScrollbarTrackEndInset(v) {
        this.i.kn = +v;
    }
    /**
     * Gets or sets whether to use automatic track insets for the vertical scrollbar in the component, when enabled.
    */
    get verticalViewScrollbarShouldAddAutoTrackInsets() {
        return this.i.hl;
    }
    set verticalViewScrollbarShouldAddAutoTrackInsets(v) {
        this.i.hl = ensureBool(v);
    }
    /**
     * Sets or gets the minimum width that the window rect is allowed to reach before being clamped.
     * Decrease this value if you want to allow for further zooming into the viewer.
     * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
     * This property accepts values between 0.0 (max zooming) and 1.0 (no zooming)
    */
    get windowRectMinWidth() {
        return this.i.kr;
    }
    set windowRectMinWidth(v) {
        this.i.kr = +v;
    }
    /**
     * Sets or gets the minimum height that the window rect is allowed to reach before being clamped.
     * Decrease this value if you want to allow for further zooming into the viewer.
     * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
     * This property accepts values between 0.0 (max zooming) and 1.0 (no zooming)
    */
    get windowRectMinHeight() {
        return this.i.kq;
    }
    set windowRectMinHeight(v) {
        this.i.kq = +v;
    }
    /**
     * Sets or gets minimum pixel width that the window is allowed to reach before being clamped.
     * Decrease this value if you want to allow for further zooming into the viewer.
     * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
     * This property is overridden by the WindowRectMinWidth property
    */
    get windowSizeMinWidth() {
        return this.i.kt;
    }
    set windowSizeMinWidth(v) {
        this.i.kt = +v;
    }
    /**
     * Sets or gets minimum pixel height that the window is allowed to reach before being clamped.
     * Decrease this value if you want to allow for further zooming into the viewer.
     * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
     * This property is overridden by the WindowRectMinHeight property
    */
    get windowSizeMinHeight() {
        return this.i.ks;
    }
    set windowSizeMinHeight(v) {
        this.i.ks = +v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.dataToolTipValueFormatSpecifiers != null && arrayFindByName(this.dataToolTipValueFormatSpecifiers, name)) {
            return arrayFindByName(this.dataToolTipValueFormatSpecifiers, name);
        }
        if (this.dataToolTipHeaderFormatSpecifiers != null && arrayFindByName(this.dataToolTipHeaderFormatSpecifiers, name)) {
            return arrayFindByName(this.dataToolTipHeaderFormatSpecifiers, name);
        }
        if (this.sortDescriptions != null && this.sortDescriptions.findByName && this.sortDescriptions.findByName(name)) {
            return this.sortDescriptions.findByName(name);
        }
        if (this.groupSortDescriptions != null && this.groupSortDescriptions.findByName && this.groupSortDescriptions.findByName(name)) {
            return this.groupSortDescriptions.findByName(name);
        }
        if (this.groupDescriptions != null && this.groupDescriptions.findByName && this.groupDescriptions.findByName(name)) {
            return this.groupDescriptions.findByName(name);
        }
        if (this.filterExpressions != null && this.filterExpressions.findByName && this.filterExpressions.findByName(name)) {
            return this.filterExpressions.findByName(name);
        }
        if (this.highlightFilterExpressions != null && this.highlightFilterExpressions.findByName && this.highlightFilterExpressions.findByName(name)) {
            return this.highlightFilterExpressions.findByName(name);
        }
        if (this.summaryDescriptions != null && this.summaryDescriptions.findByName && this.summaryDescriptions.findByName(name)) {
            return this.summaryDescriptions.findByName(name);
        }
        if (this.selectedSeriesItems != null && this.selectedSeriesItems.findByName && this.selectedSeriesItems.findByName(name)) {
            return this.selectedSeriesItems.findByName(name);
        }
        if (this.focusedSeriesItems != null && this.focusedSeriesItems.findByName && this.focusedSeriesItems.findByName(name)) {
            return this.focusedSeriesItems.findByName(name);
        }
        if (this.initialSortDescriptions != null && this.initialSortDescriptions.findByName && this.initialSortDescriptions.findByName(name)) {
            return this.initialSortDescriptions.findByName(name);
        }
        if (this.initialGroupSortDescriptions != null && this.initialGroupSortDescriptions.findByName && this.initialGroupSortDescriptions.findByName(name)) {
            return this.initialGroupSortDescriptions.findByName(name);
        }
        if (this.initialGroupDescriptions != null && this.initialGroupDescriptions.findByName && this.initialGroupDescriptions.findByName(name)) {
            return this.initialGroupDescriptions.findByName(name);
        }
        if (this.initialFilterExpressions != null && this.initialFilterExpressions.findByName && this.initialFilterExpressions.findByName(name)) {
            return this.initialFilterExpressions.findByName(name);
        }
        if (this.initialHighlightFilterExpressions != null && this.initialHighlightFilterExpressions.findByName && this.initialHighlightFilterExpressions.findByName(name)) {
            return this.initialHighlightFilterExpressions.findByName(name);
        }
        if (this.initialSummaryDescriptions != null && this.initialSummaryDescriptions.findByName && this.initialSummaryDescriptions.findByName(name)) {
            return this.initialSummaryDescriptions.findByName(name);
        }
        if (this.legend && this.legend.name && this.legend.name == name) {
            return this.legend;
        }
        if (this.markerTypes != null && this.markerTypes.findByName && this.markerTypes.findByName(name)) {
            return this.markerTypes.findByName(name);
        }
        if (this.calloutsAllowedPositions != null && this.calloutsAllowedPositions.findByName && this.calloutsAllowedPositions.findByName(name)) {
            return this.calloutsAllowedPositions.findByName(name);
        }
        if (this.valueLines != null && this.valueLines.findByName && this.valueLines.findByName(name)) {
            return this.valueLines.findByName(name);
        }
        return null;
    }
    get hasUserValues() {
        return this._hasUserValues;
    }
    __m(propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    }
    _styling(container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        let genericPrefix = "";
        let typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("DomainChart");
        let additionalPrefixes = [];
        let prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        let b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            let basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            let parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            let parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    }
    /**
 * Provides a container for this control

*
* `provideContainer` method provide container for the control.
*
* ```ts
*     this.chart.provideContainer("Container");
* ```
*/
    provideContainer(container) {
        this.i.provideContainer(container);
    }
    /**
     * Destroy this control
    
    *
    * Use the `destroy` property to destroy the chart.
    *
    * ```ts
    * this.chart.destroy();
    * ```
    */
    destroy() {
        this.i.destroy();
    }
    /**
     * Called when the control has been resized.
    
    *
    * `NotifyResized` method Called when the control has been resized.
    *
    * ```ts
    * this.chart.notifyResized();
    * ```
    */
    notifyResized() {
        this.i.notifyResized();
    }
    notifyVisualPropertiesChanged() {
        this.i.vi();
    }
    /**
     * Forces any pending refresh to the chart to be finished.
    
    *
    * Method `flush` used to Forces any pending deferred work to render on the chart before continuing.
    *
    * ```ts
    * this.chart.flush();
    * ```
    */
    flush() {
        this.i.un();
    }
    /**
     * Hides the active main tooltip, if displayed.
    
    *
    * Use the `HideToolTip` method to hide the active tooltip.
    *
    * ```ts
    * this.chart.hideToolTip();
    * ```
    */
    hideToolTip() {
        this.i.uw();
    }
    /**
     * Replays transition-in animation on all series in the chart.
    
    */
    replayTransitionIn() {
        this.i.wm();
    }
    /**
     * Performs zoom in action on the chart by specified percentage of current window rect
    
    */
    zoomIn(v) {
        this.i.zoomIn(v);
    }
    /**
     * Performs zoom out action on the chart by specified percentage of current window rect
    
    */
    zoomOut(v) {
        this.i.zoomOut(v);
    }
    /**
     * Resets the zoom level to default.
    
    */
    resetZoom() {
        this.i.resetZoom();
    }
    getDesiredToolbarActions() {
        let iv = this.i.getDesiredToolbarActions();
        return (iv);
    }
    addCommandAvailabilityListener(listener) {
        this.i.addCommandAvailabilityListener(listener);
    }
    removeCommandAvailabilityListener(listener) {
        this.i.removeCommandAvailabilityListener(listener);
    }
    addCommandStateChangedListener(listener) {
        this.i.addCommandStateChangedListener(listener);
    }
    removeCommandStateChangedListener(listener) {
        this.i.removeCommandStateChangedListener(listener);
    }
    notifySetItem(source_, index, oldItem, newItem) {
        this.i.vh(source_, index, oldItem, newItem);
    }
    /**
     * Used to manually notify the chart that the data source has reset or cleared its items.
    
    *
    * Use `notifyClearItems` method to notifies the chart that the items have been cleared from an associated data source.
    *
    * ```ts
    * this.chart.notifyClearItems(this.data);
    * ```
    */
    notifyClearItems(source_) {
        this.i.vd(source_);
    }
    notifyInsertItem(source_, index, newItem) {
        this.i.ve(source_, index, newItem);
    }
    notifyRemoveItem(source_, index, oldItem) {
        this.i.vf(source_, index, oldItem);
    }
    /**
     * Event raised when a series is initialized and added to this chart.
    *
    * `SeriesAdded` event raised when a series is initialized and added to this chart.
    *
    * ````ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	markerTypes={this.state.markersTypes}
    * 	dataSource={this.categoryData}
    * 	seriesAdded={this.CategoryChart_SeriesAdded}/>
    * ``
    *
    * ```ts
    * CategoryChart_SeriesAdded(sender:any ,aregs:ChartSeriesEventArgs)
    * {
    * }
    * ````
    */
    get seriesAdded() {
        return this._seriesAdded;
    }
    set seriesAdded(ev) {
        if (this._seriesAdded_wrapped !== null) {
            this.i.seriesAdded = delegateRemove(this.i.seriesAdded, this._seriesAdded_wrapped);
            this._seriesAdded_wrapped = null;
            this._seriesAdded = null;
        }
        this._seriesAdded = ev;
        this._seriesAdded_wrapped = (o, e) => {
            let outerArgs = new IgrChartSeriesEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSeriesAdded) {
                this.beforeSeriesAdded(this, outerArgs);
            }
            if (this._seriesAdded) {
                this._seriesAdded(this, outerArgs);
            }
        };
        this.i.seriesAdded = delegateCombine(this.i.seriesAdded, this._seriesAdded_wrapped);
        ;
    }
    /**
     * Event raised when a series is removed from this chart.
    *
    * `SeriesRemoved` event raised when a series is removed from this chart.
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	markerTypes={this.state.markersTypes}
    * 	dataSource={this.categoryData}
    * 	seriesRemoved={this.CategoryChart_seriesRemoved}/>
    * ```
    *
    * ```ts
    *  CategoryChart_seriesRemoved(sender: any , aregs: ChartSeriesEventArgs)
    *   {
    *
    *   }
    * ```
    */
    get seriesRemoved() {
        return this._seriesRemoved;
    }
    set seriesRemoved(ev) {
        if (this._seriesRemoved_wrapped !== null) {
            this.i.seriesRemoved = delegateRemove(this.i.seriesRemoved, this._seriesRemoved_wrapped);
            this._seriesRemoved_wrapped = null;
            this._seriesRemoved = null;
        }
        this._seriesRemoved = ev;
        this._seriesRemoved_wrapped = (o, e) => {
            let outerArgs = new IgrChartSeriesEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSeriesRemoved) {
                this.beforeSeriesRemoved(this, outerArgs);
            }
            if (this._seriesRemoved) {
                this._seriesRemoved(this, outerArgs);
            }
        };
        this.i.seriesRemoved = delegateCombine(this.i.seriesRemoved, this._seriesRemoved_wrapped);
        ;
    }
    /**
     * Occurs when the pointer enters a Series.
    *
    * `SeriesPointerEnter` event raised when the pointer enters a Series.
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	markerTypes={this.state.markersTypes}
    * 	dataSource={this.categoryData}
    * 	seriesPointerEnter={this.CategoryChart_seriesPointerDown}/>
    * ```
    *
    * ```ts
    * CategoryChart_seriesPointerEnter(sender: any , aregs: ChartSeriesEventArgs)
    *   {
    *
    *   }
    * ```
    */
    get seriesPointerEnter() {
        return this._seriesPointerEnter;
    }
    set seriesPointerEnter(ev) {
        if (this._seriesPointerEnter_wrapped !== null) {
            this.i.seriesPointerEnter = delegateRemove(this.i.seriesPointerEnter, this._seriesPointerEnter_wrapped);
            this._seriesPointerEnter_wrapped = null;
            this._seriesPointerEnter = null;
        }
        this._seriesPointerEnter = ev;
        this._seriesPointerEnter_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartSeriesPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSeriesPointerEnter) {
                this.beforeSeriesPointerEnter(this, outerArgs);
            }
            if (this._seriesPointerEnter) {
                this._seriesPointerEnter(this, outerArgs);
            }
        };
        this.i.seriesPointerEnter = delegateCombine(this.i.seriesPointerEnter, this._seriesPointerEnter_wrapped);
        ;
    }
    /**
     * Occurs when the pointer leaves a Series.
    *
    * `SeriesPointerLeave` event raised when the pointer leaves a Series.
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	markerTypes={this.state.markersTypes}
    * 	dataSource={this.categoryData}
    * 	seriesPointerLeave={this.CategoryChart_seriesPointerLeave}/>
    * ```
    *
    * ```ts
    * CategoryChart_seriesPointerLeave(sender: any , aregs: ChartSeriesEventArgs)
    *   {
    *
    *   }
    * ```
    */
    get seriesPointerLeave() {
        return this._seriesPointerLeave;
    }
    set seriesPointerLeave(ev) {
        if (this._seriesPointerLeave_wrapped !== null) {
            this.i.seriesPointerLeave = delegateRemove(this.i.seriesPointerLeave, this._seriesPointerLeave_wrapped);
            this._seriesPointerLeave_wrapped = null;
            this._seriesPointerLeave = null;
        }
        this._seriesPointerLeave = ev;
        this._seriesPointerLeave_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartSeriesPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSeriesPointerLeave) {
                this.beforeSeriesPointerLeave(this, outerArgs);
            }
            if (this._seriesPointerLeave) {
                this._seriesPointerLeave(this, outerArgs);
            }
        };
        this.i.seriesPointerLeave = delegateCombine(this.i.seriesPointerLeave, this._seriesPointerLeave_wrapped);
        ;
    }
    /**
     * Occurs when the pointer moves over a Series.
    *
    * `SeriesPointerMove` event raised when the pointer moves over a Series.
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	markerTypes={this.state.markersTypes}
    * 	dataSource={this.categoryData}
    * 	seriesPointerMove={this.CategoryChart_seriesPointerMove}/>
    * ```
    *
    * ```ts
    *  CategoryChart_seriesPointerMove(sender: any , aregs: ChartSeriesEventArgs)
    *   {
    *
    *   }
    * ```
    */
    get seriesPointerMove() {
        return this._seriesPointerMove;
    }
    set seriesPointerMove(ev) {
        if (this._seriesPointerMove_wrapped !== null) {
            this.i.seriesPointerMove = delegateRemove(this.i.seriesPointerMove, this._seriesPointerMove_wrapped);
            this._seriesPointerMove_wrapped = null;
            this._seriesPointerMove = null;
        }
        this._seriesPointerMove = ev;
        this._seriesPointerMove_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartSeriesPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSeriesPointerMove) {
                this.beforeSeriesPointerMove(this, outerArgs);
            }
            if (this._seriesPointerMove) {
                this._seriesPointerMove(this, outerArgs);
            }
        };
        this.i.seriesPointerMove = delegateCombine(this.i.seriesPointerMove, this._seriesPointerMove_wrapped);
        ;
    }
    /**
     * Occurs when the pointer is pressed down over a Series.
    *
    * `SeriesPointerDown` event raised when the pointer is pressed down over a Series
    *
    * ````ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	markerTypes={this.state.markersTypes}
    * 	dataSource={this.categoryData}
    * 	seriesPointerDown={this.CategoryChart_seriesPointerDown}/>
    * ``
    *
    * ```ts
    * CategoryChart_seriesPointerDown(sender: any , aregs: ChartSeriesEventArgs)
    *   {
    *
    *   }
    * ````
    */
    get seriesPointerDown() {
        return this._seriesPointerDown;
    }
    set seriesPointerDown(ev) {
        if (this._seriesPointerDown_wrapped !== null) {
            this.i.seriesPointerDown = delegateRemove(this.i.seriesPointerDown, this._seriesPointerDown_wrapped);
            this._seriesPointerDown_wrapped = null;
            this._seriesPointerDown = null;
        }
        this._seriesPointerDown = ev;
        this._seriesPointerDown_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartSeriesPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSeriesPointerDown) {
                this.beforeSeriesPointerDown(this, outerArgs);
            }
            if (this._seriesPointerDown) {
                this._seriesPointerDown(this, outerArgs);
            }
        };
        this.i.seriesPointerDown = delegateCombine(this.i.seriesPointerDown, this._seriesPointerDown_wrapped);
        ;
    }
    /**
     * Occurs when the pointer is released over a Series.
    *
    * `SeriesPointerUp` event raised when the pointer is released over a Series.
    *
    * ```ts
    *  <IgrCategoryChart
    * 	width="100%"
    * 	height="100%"
    * 	chartType="Line"
    * 	markerTypes={this.state.markersTypes}
    * 	dataSource={this.categoryData}
    * 	seriesPointerUp={this.CategoryChart_seriesPointerUp}/>
    * ```
    *
    * ```ts
    *  CategoryChart_seriesPointerUp(sender: any , aregs: ChartSeriesEventArgs)
    *   {
    *
    *   }
    * ```
    */
    get seriesPointerUp() {
        return this._seriesPointerUp;
    }
    set seriesPointerUp(ev) {
        if (this._seriesPointerUp_wrapped !== null) {
            this.i.seriesPointerUp = delegateRemove(this.i.seriesPointerUp, this._seriesPointerUp_wrapped);
            this._seriesPointerUp_wrapped = null;
            this._seriesPointerUp = null;
        }
        this._seriesPointerUp = ev;
        this._seriesPointerUp_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartSeriesPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSeriesPointerUp) {
                this.beforeSeriesPointerUp(this, outerArgs);
            }
            if (this._seriesPointerUp) {
                this._seriesPointerUp(this, outerArgs);
            }
        };
        this.i.seriesPointerUp = delegateCombine(this.i.seriesPointerUp, this._seriesPointerUp_wrapped);
        ;
    }
    /**
     * Occurs when the pointer is clicked over a series.
    */
    get seriesClick() {
        return this._seriesClick;
    }
    set seriesClick(ev) {
        if (this._seriesClick_wrapped !== null) {
            this.i.seriesClick = delegateRemove(this.i.seriesClick, this._seriesClick_wrapped);
            this._seriesClick_wrapped = null;
            this._seriesClick = null;
        }
        this._seriesClick = ev;
        this._seriesClick_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartSeriesPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSeriesClick) {
                this.beforeSeriesClick(this, outerArgs);
            }
            if (this._seriesClick) {
                this._seriesClick(this, outerArgs);
            }
        };
        this.i.seriesClick = delegateCombine(this.i.seriesClick, this._seriesClick_wrapped);
        ;
    }
    /**
     * Occurs when the pointer enters the plot area.
    */
    get plotAreaPointerEnter() {
        return this._plotAreaPointerEnter;
    }
    set plotAreaPointerEnter(ev) {
        if (this._plotAreaPointerEnter_wrapped !== null) {
            this.i.plotAreaPointerEnter = delegateRemove(this.i.plotAreaPointerEnter, this._plotAreaPointerEnter_wrapped);
            this._plotAreaPointerEnter_wrapped = null;
            this._plotAreaPointerEnter = null;
        }
        this._plotAreaPointerEnter = ev;
        this._plotAreaPointerEnter_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartPlotAreaPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforePlotAreaPointerEnter) {
                this.beforePlotAreaPointerEnter(this, outerArgs);
            }
            if (this._plotAreaPointerEnter) {
                this._plotAreaPointerEnter(this, outerArgs);
            }
        };
        this.i.plotAreaPointerEnter = delegateCombine(this.i.plotAreaPointerEnter, this._plotAreaPointerEnter_wrapped);
        ;
    }
    /**
     * Occurs when the pointer leaves the plot area.
    */
    get plotAreaPointerLeave() {
        return this._plotAreaPointerLeave;
    }
    set plotAreaPointerLeave(ev) {
        if (this._plotAreaPointerLeave_wrapped !== null) {
            this.i.plotAreaPointerLeave = delegateRemove(this.i.plotAreaPointerLeave, this._plotAreaPointerLeave_wrapped);
            this._plotAreaPointerLeave_wrapped = null;
            this._plotAreaPointerLeave = null;
        }
        this._plotAreaPointerLeave = ev;
        this._plotAreaPointerLeave_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartPlotAreaPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforePlotAreaPointerLeave) {
                this.beforePlotAreaPointerLeave(this, outerArgs);
            }
            if (this._plotAreaPointerLeave) {
                this._plotAreaPointerLeave(this, outerArgs);
            }
        };
        this.i.plotAreaPointerLeave = delegateCombine(this.i.plotAreaPointerLeave, this._plotAreaPointerLeave_wrapped);
        ;
    }
    /**
     * Occurs when the pointer is over the plot area.
    */
    get plotAreaPointerMove() {
        return this._plotAreaPointerMove;
    }
    set plotAreaPointerMove(ev) {
        if (this._plotAreaPointerMove_wrapped !== null) {
            this.i.plotAreaPointerMove = delegateRemove(this.i.plotAreaPointerMove, this._plotAreaPointerMove_wrapped);
            this._plotAreaPointerMove_wrapped = null;
            this._plotAreaPointerMove = null;
        }
        this._plotAreaPointerMove = ev;
        this._plotAreaPointerMove_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartPlotAreaPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforePlotAreaPointerMove) {
                this.beforePlotAreaPointerMove(this, outerArgs);
            }
            if (this._plotAreaPointerMove) {
                this._plotAreaPointerMove(this, outerArgs);
            }
        };
        this.i.plotAreaPointerMove = delegateCombine(this.i.plotAreaPointerMove, this._plotAreaPointerMove_wrapped);
        ;
    }
    /**
     * Occurs when the pointer is pressed down over the plot area.
    */
    get plotAreaPointerDown() {
        return this._plotAreaPointerDown;
    }
    set plotAreaPointerDown(ev) {
        if (this._plotAreaPointerDown_wrapped !== null) {
            this.i.plotAreaPointerDown = delegateRemove(this.i.plotAreaPointerDown, this._plotAreaPointerDown_wrapped);
            this._plotAreaPointerDown_wrapped = null;
            this._plotAreaPointerDown = null;
        }
        this._plotAreaPointerDown = ev;
        this._plotAreaPointerDown_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartPlotAreaPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforePlotAreaPointerDown) {
                this.beforePlotAreaPointerDown(this, outerArgs);
            }
            if (this._plotAreaPointerDown) {
                this._plotAreaPointerDown(this, outerArgs);
            }
        };
        this.i.plotAreaPointerDown = delegateCombine(this.i.plotAreaPointerDown, this._plotAreaPointerDown_wrapped);
        ;
    }
    /**
     * Occurs when the pointer is released over the plot area.
    */
    get plotAreaPointerUp() {
        return this._plotAreaPointerUp;
    }
    set plotAreaPointerUp(ev) {
        if (this._plotAreaPointerUp_wrapped !== null) {
            this.i.plotAreaPointerUp = delegateRemove(this.i.plotAreaPointerUp, this._plotAreaPointerUp_wrapped);
            this._plotAreaPointerUp_wrapped = null;
            this._plotAreaPointerUp = null;
        }
        this._plotAreaPointerUp = ev;
        this._plotAreaPointerUp_wrapped = (o, e) => {
            let outerArgs = new IgrDomainChartPlotAreaPointerEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforePlotAreaPointerUp) {
                this.beforePlotAreaPointerUp(this, outerArgs);
            }
            if (this._plotAreaPointerUp) {
                this._plotAreaPointerUp(this, outerArgs);
            }
        };
        this.i.plotAreaPointerUp = delegateCombine(this.i.plotAreaPointerUp, this._plotAreaPointerUp_wrapped);
        ;
    }
    /**
     * Occurs when the style of a callout is updated.
    */
    get calloutStyleUpdating() {
        return this._calloutStyleUpdating;
    }
    set calloutStyleUpdating(ev) {
        if (this._calloutStyleUpdating_wrapped !== null) {
            this.i.calloutStyleUpdating = delegateRemove(this.i.calloutStyleUpdating, this._calloutStyleUpdating_wrapped);
            this._calloutStyleUpdating_wrapped = null;
            this._calloutStyleUpdating = null;
        }
        this._calloutStyleUpdating = ev;
        this._calloutStyleUpdating_wrapped = (o, e) => {
            let outerArgs = new IgrCalloutStyleUpdatingEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeCalloutStyleUpdating) {
                this.beforeCalloutStyleUpdating(this, outerArgs);
            }
            if (this._calloutStyleUpdating) {
                this._calloutStyleUpdating(this, outerArgs);
            }
        };
        this.i.calloutStyleUpdating = delegateCombine(this.i.calloutStyleUpdating, this._calloutStyleUpdating_wrapped);
        ;
    }
    /**
     * Occurs when the style of a callout is updated.
    */
    get calloutRenderStyleUpdating() {
        return this._calloutRenderStyleUpdating;
    }
    set calloutRenderStyleUpdating(ev) {
        if (this._calloutRenderStyleUpdating_wrapped !== null) {
            this.i.calloutRenderStyleUpdating = delegateRemove(this.i.calloutRenderStyleUpdating, this._calloutRenderStyleUpdating_wrapped);
            this._calloutRenderStyleUpdating_wrapped = null;
            this._calloutRenderStyleUpdating = null;
        }
        this._calloutRenderStyleUpdating = ev;
        this._calloutRenderStyleUpdating_wrapped = (o, e) => {
            let outerArgs = new IgrCalloutRenderStyleUpdatingEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeCalloutRenderStyleUpdating) {
                this.beforeCalloutRenderStyleUpdating(this, outerArgs);
            }
            if (this._calloutRenderStyleUpdating) {
                this._calloutRenderStyleUpdating(this, outerArgs);
            }
        };
        this.i.calloutRenderStyleUpdating = delegateCombine(this.i.calloutRenderStyleUpdating, this._calloutRenderStyleUpdating_wrapped);
        ;
    }
    get selectedSeriesItemsChanged() {
        return this._selectedSeriesItemsChanged;
    }
    set selectedSeriesItemsChanged(ev) {
        if (this._selectedSeriesItemsChanged_wrapped !== null) {
            this.i.selectedSeriesItemsChanged = delegateRemove(this.i.selectedSeriesItemsChanged, this._selectedSeriesItemsChanged_wrapped);
            this._selectedSeriesItemsChanged_wrapped = null;
            this._selectedSeriesItemsChanged = null;
        }
        this._selectedSeriesItemsChanged = ev;
        this._selectedSeriesItemsChanged_wrapped = (o, e) => {
            let outerArgs = new IgrSeriesViewerSelectedSeriesItemsChangedEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSelectedSeriesItemsChanged) {
                this.beforeSelectedSeriesItemsChanged(this, outerArgs);
            }
            if (this._selectedSeriesItemsChanged) {
                this._selectedSeriesItemsChanged(this, outerArgs);
            }
        };
        this.i.selectedSeriesItemsChanged = delegateCombine(this.i.selectedSeriesItemsChanged, this._selectedSeriesItemsChanged_wrapped);
        ;
    }
    get focusedSeriesItemsChanged() {
        return this._focusedSeriesItemsChanged;
    }
    set focusedSeriesItemsChanged(ev) {
        if (this._focusedSeriesItemsChanged_wrapped !== null) {
            this.i.focusedSeriesItemsChanged = delegateRemove(this.i.focusedSeriesItemsChanged, this._focusedSeriesItemsChanged_wrapped);
            this._focusedSeriesItemsChanged_wrapped = null;
            this._focusedSeriesItemsChanged = null;
        }
        this._focusedSeriesItemsChanged = ev;
        this._focusedSeriesItemsChanged_wrapped = (o, e) => {
            let outerArgs = new IgrSeriesViewerSelectedSeriesItemsChangedEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeFocusedSeriesItemsChanged) {
                this.beforeFocusedSeriesItemsChanged(this, outerArgs);
            }
            if (this._focusedSeriesItemsChanged) {
                this._focusedSeriesItemsChanged(this, outerArgs);
            }
        };
        this.i.focusedSeriesItemsChanged = delegateCombine(this.i.focusedSeriesItemsChanged, this._focusedSeriesItemsChanged_wrapped);
        ;
    }
    get filterStringErrorsParsing() {
        return this._filterStringErrorsParsing;
    }
    set filterStringErrorsParsing(ev) {
        if (this._filterStringErrorsParsing_wrapped !== null) {
            this.i.filterStringErrorsParsing = delegateRemove(this.i.filterStringErrorsParsing, this._filterStringErrorsParsing_wrapped);
            this._filterStringErrorsParsing_wrapped = null;
            this._filterStringErrorsParsing = null;
        }
        this._filterStringErrorsParsing = ev;
        this._filterStringErrorsParsing_wrapped = (o, e) => {
            let outerArgs = new IgrFilterStringErrorsParsingEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeFilterStringErrorsParsing) {
                this.beforeFilterStringErrorsParsing(this, outerArgs);
            }
            if (this._filterStringErrorsParsing) {
                this._filterStringErrorsParsing(this, outerArgs);
            }
        };
        this.i.filterStringErrorsParsing = delegateCombine(this.i.filterStringErrorsParsing, this._filterStringErrorsParsing_wrapped);
        ;
    }
}
