// igniteui-react-charts/esm2015/lib/igr-funnel-slice-clicked-event-args.js
import { FunnelSliceClickedEventArgs as FunnelSliceClickedEventArgs_internal } from "./FunnelSliceClickedEventArgs";
import { fromRect, toRect } from "igniteui-react-core";
/**
 * Information about the clicked slice.
*/
export class IgrFunnelSliceClickedEventArgs {
    createImplementation() {
        return new FunnelSliceClickedEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * The index of the clicked slice.
    */
    get index() {
        return this.i.index;
    }
    set index(v) {
        this.i.index = +v;
    }
    /**
     * The bounds of the slice.
    */
    get bounds() {
        return fromRect(this.i.bounds);
    }
    set bounds(v) {
        this.i.bounds = toRect(v);
    }
    /**
     * The data item for the clicked slice.
    */
    get item() {
        return this.i.item;
    }
    set item(v) {
        this.i.item = v;
    }
}
