// igniteui-react-charts/esm2015/lib/igr-annotation-layer.js
import { AnnotationAppearanceMode_$type } from "./AnnotationAppearanceMode";
import { IgrSeries } from "./igr-series";
import { ensureBool, fromPoint, toPoint, ensureEnum, toDoubleCollection, fromDoubleCollection } from "igniteui-react-core";
/**
 * Represents the base class for annotation layer types. Annotations will usually disable some of the default hover/touch behaviors of the chart when they are introduced to the series collection.
 * In return they provide some valuable information as the cursor is moved by hovering over the chart, or performing a press-hold-drag interaction in touch mode.
*/
export class IgrAnnotationLayer extends IgrSeries {
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or sets whether this series should take up an index for auto brush coloring.
    */
    get useIndex() {
        return this.i.xk;
    }
    set useIndex(v) {
        this.i.xk = ensureBool(v);
    }
    /**
     * Gets or sets whether this series interacts with the chart legend.
    */
    get useLegend() {
        return this.i.xl;
    }
    set useLegend(v) {
        this.i.xl = ensureBool(v);
    }
    /**
     * Gets or sets the world position to use instead of the pointer position for this annotation layer. If you set this value, rather than responding to the pointer, this layer will be fixed on the provided world position.
     * A world position has x and y values that range from 0 to 1 that represent the global position of the cursor relative to the entire ranges of the axes. The default of (NaN, NaN) means that the layer should react to pointer movement.
    */
    get cursorPosition() {
        return fromPoint(this.i.zv);
    }
    set cursorPosition(v) {
        this.i.zv = toPoint(v);
    }
    /**
     * Gets or sets whether the series viewer's default crosshair should be disabled by the presence of this layer.
    */
    get isDefaultCrosshairDisabled() {
        return this.i.xf;
    }
    set isDefaultCrosshairDisabled(v) {
        this.i.xf = ensureBool(v);
    }
    /**
     * Gets whether or not the default crosshair behavior is disabled.
    */
    get isDefaultCrosshairBehaviorDisabled() {
        return this.i.e1;
    }
    /**
     * Gets or sets the mode used for shifting the annotation layer color based on the target series.
    */
    get appearanceMode() {
        return this.i.w1;
    }
    set appearanceMode(v) {
        this.i.w1 = ensureEnum(AnnotationAppearanceMode_$type, v);
    }
    /**
     * Gets or sets the mode used for shifting the annotation layer color based on the target series.
    */
    get actualAppearanceMode() {
        return this.i.wy;
    }
    set actualAppearanceMode(v) {
        this.i.wy = ensureEnum(AnnotationAppearanceMode_$type, v);
    }
    /**
     * Gets or sets the percentage amount to shift the appearance mode by. Value must range between -1.0 to 1.0.
    */
    get shiftAmount() {
        return this.i.xy;
    }
    set shiftAmount(v) {
        this.i.xy = +v;
    }
    /**
     * Gets or sets the percentage amount to shift the appearance mode by. Value must range between -1.0 to 1.0.
    */
    get actualShiftAmount() {
        return this.i.xn;
    }
    set actualShiftAmount(v) {
        this.i.xn = +v;
    }
    /**
     * Gets or sets the mode used for shifting the annotation layer color based on the target series.
    */
    get horizontalAppearanceMode() {
        return this.i.w8;
    }
    set horizontalAppearanceMode(v) {
        this.i.w8 = ensureEnum(AnnotationAppearanceMode_$type, v);
    }
    /**
     * Gets or sets the mode used for shifting the annotation layer color based on the target series.
    */
    get actualHorizontalAppearanceMode() {
        return this.i.wz;
    }
    set actualHorizontalAppearanceMode(v) {
        this.i.wz = ensureEnum(AnnotationAppearanceMode_$type, v);
    }
    /**
     * Gets or sets the mode used for shifting the annotation layer color based on the target series.
    */
    get verticalAppearanceMode() {
        return this.i.w9;
    }
    set verticalAppearanceMode(v) {
        this.i.w9 = ensureEnum(AnnotationAppearanceMode_$type, v);
    }
    /**
     * Gets or sets the mode used for shifting the annotation layer color based on the target series.
    */
    get actualVerticalAppearanceMode() {
        return this.i.w0;
    }
    set actualVerticalAppearanceMode(v) {
        this.i.w0 = ensureEnum(AnnotationAppearanceMode_$type, v);
    }
    /**
     * Gets or sets the percentage amount to shift the appearance mode by. Value must range between -1.0 to 1.0.
    */
    get horizontalShiftAmount() {
        return this.i.xx;
    }
    set horizontalShiftAmount(v) {
        this.i.xx = +v;
    }
    /**
     * Gets or sets the percentage amount to shift the appearance mode by. Value must range between -1.0 to 1.0.
    */
    get actualHorizontalShiftAmount() {
        return this.i.xm;
    }
    set actualHorizontalShiftAmount(v) {
        this.i.xm = +v;
    }
    /**
     * Gets or sets the percentage amount to shift the appearance mode by. Value must range between -1.0 to 1.0.
    */
    get verticalShiftAmount() {
        return this.i.xz;
    }
    set verticalShiftAmount(v) {
        this.i.xz = +v;
    }
    /**
     * Gets or sets the percentage amount to shift the appearance mode by. Value must range between -1.0 to 1.0.
    */
    get actualVerticalShiftAmount() {
        return this.i.xo;
    }
    set actualVerticalShiftAmount(v) {
        this.i.xo = +v;
    }
    /**
     * Gets or sets the mode used for shifting the annotation layer color based on the target series.
    */
    get horizontalDashArray() {
        return fromDoubleCollection(this.i.zs);
    }
    set horizontalDashArray(v) {
        this.i.zs = toDoubleCollection(v);
    }
    /**
     * Gets or sets the mode used for shifting the annotation layer color based on the target series.
    */
    get actualHorizontalDashArray() {
        return fromDoubleCollection(this.i.zm);
    }
    set actualHorizontalDashArray(v) {
        this.i.zm = toDoubleCollection(v);
    }
    /**
     * Gets or sets the mode used for shifting the annotation layer color based on the target series.
    */
    get verticalDashArray() {
        return fromDoubleCollection(this.i.zt);
    }
    set verticalDashArray(v) {
        this.i.zt = toDoubleCollection(v);
    }
    /**
     * Gets or sets the mode used for shifting the annotation layer color based on the target series.
    */
    get actualVerticalDashArray() {
        return fromDoubleCollection(this.i.zn);
    }
    set actualVerticalDashArray(v) {
        this.i.zn = toDoubleCollection(v);
    }
    /**
     * Gets if the series uses an index, which can be used to auto-assign it a color.
    */
    get isIndexed() {
        return this.i.fg;
    }
    /**
     * Gets if the series should appear in any legends.
    */
    get isUsableInLegend() {
        return this.i.isUsableInLegend;
    }
    /**
     * Gets whether this series is an annotation layer.
    */
    get isAnnotationLayer() {
        return this.i.es;
    }
    /**
     * Gets or sets whether this layer renders as an overlay or not.
    */
    get shouldRenderAsOverlay() {
        return this.i.xj;
    }
    set shouldRenderAsOverlay(v) {
        this.i.xj = ensureBool(v);
    }
}
