// igniteui-react-charts/esm2015/lib/igr-chart-cursor-event-args.js
import { IgrSeries } from "./igr-series";
/**
 * Provides data for IgxDataChartComponent mouse button related events.
*/
export class IgrChartCursorEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * Gets the ItemsSource item associated with the current event.
    */
    get item() {
        return this.i.item;
    }
    set item(v) {
        this.i.item = v;
    }
    /**
     * Gets the series associated with the current event.
    */
    get series() {
        const r = this.i.series;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrSeries._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set series(v) {
        v == null ? this.i.series = null : this.i.series = v.i;
    }
    /**
     * Gets the Chart associated with the current event.
    */
    get seriesViewer() {
        const r = this.i.seriesViewer;
        if (r == null) {
            return null;
        }
        return r.externalObject;
    }
    set seriesViewer(v) {
        v == null ? this.i.seriesViewer = null : this.i.seriesViewer = v.i;
    }
    /**
     * Provides a human readable expresion of the event arguments.
    
    */
    toString() {
        let iv = this.i.toString();
        return (iv);
    }
}
