// igniteui-react-charts/esm2015/lib/igr-transition-out-completed-event-args.js
import { TransitionOutCompletedEventArgs as TransitionOutCompletedEventArgs_internal } from "./TransitionOutCompletedEventArgs";
/**
 * EventArgs class for TransitionOutCompleted events.
*/
export class IgrTransitionOutCompletedEventArgs {
    createImplementation() {
        return new TransitionOutCompletedEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
}
