// igniteui-react-charts/esm2015/lib/igr-domain-chart-plot-area-pointer-event-args.js
import { fromPoint, toPoint } from "igniteui-react-core";
/**
 * Represents event arguments for chart's SeriesAdded and SeriesRemoved events
*/
export class IgrDomainChartPlotAreaPointerEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * Gets the pointer position relative to the plot area.
    */
    get plotAreaPosition() {
        return fromPoint(this.i.plotAreaPosition);
    }
    set plotAreaPosition(v) {
        this.i.plotAreaPosition = toPoint(v);
    }
    /**
     * Gets the mouse position relative to the chart.
    */
    get chartPosition() {
        return fromPoint(this.i.chartPosition);
    }
}
