// igniteui-react-charts/esm2015/lib/igr-stacked-column-series.js
import { IgrHorizontalStackedSeriesBase } from "./igr-horizontal-stacked-series-base";
import { StackedColumnSeries } from "./StackedColumnSeries";
/**
 * Represents a IgxDataChartComponent stacked column series.
*/
export class IgrStackedColumnSeries extends IgrHorizontalStackedSeriesBase {
    createImplementation() {
        return new StackedColumnSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the current series shows a column shape.
    */
    get isColumn() {
        return this.i.ez;
    }
    /**
     * Overridden by derived series classes to indicate when marker-less display is preferred or not.
    */
    get isMarkerlessDisplayPreferred() {
        return this.i.fl;
    }
    /**
     * Gets or sets the x-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusX() {
        return this.i.ab3;
    }
    set radiusX(v) {
        this.i.ab3 = +v;
    }
    /**
     * Gets or sets the y-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusY() {
        return this.i.ab4;
    }
    set radiusY(v) {
        this.i.ab4 = +v;
    }
    /**
     * For a category plotted series, returns the current width of the items within the categories. This only returns a value if the items have some form of width (e.g. columns, bars, etc.) otherwise 0 is returned.
    
    */
    getItemSpan() {
        let iv = this.i.iw();
        return (iv);
    }
}
