// igniteui-react-charts/esm2015/lib/igr-filter-string-errors-parsing-event-args.js
import { FilterStringErrorsParsingEventArgs as FilterStringErrorsParsingEventArgs_internal } from "./FilterStringErrorsParsingEventArgs";
/**
 * Represents event arguments for when there are errors parsing the filter string
*/
export class IgrFilterStringErrorsParsingEventArgs {
    createImplementation() {
        return new FilterStringErrorsParsingEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get propertyName() {
        return this.i.propertyName;
    }
    set propertyName(v) {
        this.i.propertyName = v;
    }
    /**
     * Gets the errors from parsing the filter string.
    */
    get errors() {
        return this.i.errors;
    }
    set errors(v) {
        this.i.errors = v;
    }
}
