// igniteui-react-charts/esm2015/lib/igr-zoom-slider-resolving-axis-value-event-args.js
import { ZoomSliderResolvingAxisValueEventArgs as ZoomSliderResolvingAxisValueEventArgs_internal } from "./ZoomSliderResolvingAxisValueEventArgs";
export class IgrZoomSliderResolvingAxisValueEventArgs {
    createImplementation() {
        return new ZoomSliderResolvingAxisValueEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get position() {
        return this.i.position;
    }
    set position(v) {
        this.i.position = +v;
    }
    get value() {
        return this.i.value;
    }
    set value(v) {
        this.i.value = v;
    }
}
