// igniteui-react-charts/esm2015/lib/igr-assigning-series-shape-style-event-args-base.js
import { IgrAssigningSeriesStyleEventArgsBase } from "./igr-assigning-series-style-event-args-base";
import { toDoubleCollection, fromDoubleCollection } from "igniteui-react-core";
/**
 * Represents event arguments class for the AssigningCategoryStyleEvent
*/
export class IgrAssigningSeriesShapeStyleEventArgsBase extends IgrAssigningSeriesStyleEventArgsBase {
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    /**
     * Gets or sets stroke thickness to use for the current item.
    */
    get strokeThickness() {
        return this.i.ae;
    }
    set strokeThickness(v) {
        this.i.ae = +v;
    }
    /**
     * Gets or sets stroke dash array to use for the current item.
    */
    get strokeDashArray() {
        return fromDoubleCollection(this.i.af);
    }
    set strokeDashArray(v) {
        this.i.af = toDoubleCollection(v);
    }
    /**
     * Gets or sets corner radius X to use for the current item, if applicable.
    */
    get radiusX() {
        return this.i.ac;
    }
    set radiusX(v) {
        this.i.ac = +v;
    }
    /**
     * Gets or sets corner radius Y to use for the current item, if applicable.
    */
    get radiusY() {
        return this.i.ad;
    }
    set radiusY(v) {
        this.i.ad = +v;
    }
}
