// igniteui-react-charts/esm2015/lib/igr-range-area-series.js
import { IgrHorizontalRangeCategorySeries } from "./igr-horizontal-range-category-series";
import { RangeAreaSeries } from "./RangeAreaSeries";
/**
 * Represents a IgxDataChartComponent range area series.
*
* The `RangeAreaSeries` class represents the range area series of the IgxDataChartComponent.
*
* ```ts
* <IgrDataChart
*     dataSource={this.state.dataSource}
*     width="700px"
*     height="500px">
*
*     <IgrCategoryXAxis name="xAxis" label="Year" />
*     <IgrNumericYAxis  name="yAxis" />
*
*     <IgrRangeAreaSeries
*         name="series1"
*         xAxisName="xAxis"
*         yAxisName="yAxis"
*         highMemberPath="High"
*         lowMemberPath="Low" />
*  </IgrDataChart>
* ```
*
* ```ts
* let series = new IgrRangeAreaSeries({name:"series1"});
* series.xAxisName = this.xAxis;
* series.yAxisName = this.yAxis;
* series.openMemberPath = "open";
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* ```
*/
export class IgrRangeAreaSeries extends IgrHorizontalRangeCategorySeries {
    createImplementation() {
        return new RangeAreaSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the current series shows an area or line shape.
    *
    * Use the `IsAreaOrLine` property to check if the current series shows an area or line shape.
    *
    * ```ts
    * var x= series.isAreaOrLine;
    * ```
    */
    get isAreaOrLine() {
        return this.i.ev;
    }
    /**
     * Gets whether the current series shows an area shape.
    */
    get isArea() {
        return this.i.eu;
    }
}
