// igniteui-react-charts/esm2015/lib/igr-radial-line-series.js
import { UnknownValuePlotting_$type } from "igniteui-react-core";
import { IgrAnchoredRadialSeries } from "./igr-anchored-radial-series";
import { RadialLineSeries } from "./RadialLineSeries";
import { ensureEnum } from "igniteui-react-core";
/**
 * Represents a IgxDataChartComponent radial line series.
*/
export class IgrRadialLineSeries extends IgrAnchoredRadialSeries {
    createImplementation() {
        return new RadialLineSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the current series shows only line shapes.
    */
    get isLineOnly() {
        return this.i.fk;
    }
    /**
     * Determines how unknown values will be plotted on the chart.
     * Null and Double.NaN are two examples of unknown values.
    */
    get unknownValuePlotting() {
        return this.i.adq;
    }
    set unknownValuePlotting(v) {
        this.i.adq = ensureEnum(UnknownValuePlotting_$type, v);
    }
    /**
     * When overridden in a derived class, is invoked whenever application code or internal processes
     * call ApplyTemplate.
    
    */
    onApplyTemplate() {
        this.i.ac();
    }
}
