// igniteui-react-charts/esm2015/lib/igr-value-layer.js
import { ValueLayerValueMode_$type } from "./ValueLayerValueMode";
import { IgrSeries } from "./igr-series";
import { IgrAxis } from "./igr-axis";
import { IgrAnnotationLayer } from "./igr-annotation-layer";
import { ValueLayer } from "./ValueLayer";
import { ensureBool, ensureEnum, brushToString, stringToBrush, toPoint } from "igniteui-react-core";
/**
 * Represents an annotation layer that displays crosshair lines that cross through the closest value of the target series under the cursor.
*/
export class IgrValueLayer extends IgrAnnotationLayer {
    createImplementation() {
        return new ValueLayer();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the series is an crosshair annotation layer.
    */
    get isAnnotationValueLayer() {
        return this.i.et;
    }
    /**
     * Gets or sets the value mode for the overlay.
    */
    get valueMode() {
        return this.i.z5;
    }
    set valueMode(v) {
        this.i.z5 = ensureEnum(ValueLayerValueMode_$type, v);
    }
    /**
     * Gets or sets the color to use for the horizontal line. Leave null for an automatic value.
    */
    get horizontalLineStroke() {
        return brushToString(this.i.acu);
    }
    set horizontalLineStroke(v) {
        this.i.acu = stringToBrush(v);
    }
    /**
     * Gets or sets the color to use for vertical line. Leave null for an automatic value.
    */
    get verticalLineStroke() {
        return brushToString(this.i.acv);
    }
    set verticalLineStroke(v) {
        this.i.acv = stringToBrush(v);
    }
    /**
     * Gets or sets the name of the series to target this annotation to. If null, this annotation targets all series simultaneously.
    */
    get targetSeriesName() {
        return this.i.aay;
    }
    set targetSeriesName(v) {
        this.i.aay = v;
    }
    /**
     * Gets or sets the series to target this annotation to. If null, this annotation targets all series simultaneously.
    */
    get targetSeries() {
        const r = this.i.z3;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrSeries._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set targetSeries(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.z3 = null : this.i.z3 = v.i;
    }
    /**
     * Gets or sets the name of the axis to target this annotation to. If null, this annotation targets all axis simultaneously.
    */
    get targetAxisName() {
        return this.i.aav;
    }
    set targetAxisName(v) {
        this.i.aav = v;
    }
    /**
     * Gets or sets the axis to target this annotation to. If null, this annotation targets all value axis simultaneously.
    */
    get targetAxis() {
        const r = this.i.zy;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrAxis._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set targetAxis(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.zy = null : this.i.zy = v.i;
    }
    /**
     * Gets or sets whether to use value interpolation when drawing a line through the best value for the cursor position.
    */
    get useInterpolation() {
        return this.i.aaa;
    }
    set useInterpolation(v) {
        this.i.aaa = ensureBool(v);
    }
    /**
     * Gets or sets whether to draw annotations over the axes where the crosshair meets with them.
    */
    get isAxisAnnotationEnabled() {
        return this.i.z7;
    }
    set isAxisAnnotationEnabled(v) {
        this.i.z7 = ensureBool(v);
    }
    /**
     * Sets or gets a function which takes an object that produces a formatted label for the x axis annotation.
    */
    get xAxisAnnotationFormatLabel() {
        return this.i.zz;
    }
    set xAxisAnnotationFormatLabel(v) {
        this.i.zz = v;
    }
    /**
     * Sets or gets a function which takes an object that produces a formatted label for the y axis annotation.
    */
    get yAxisAnnotationFormatLabel() {
        return this.i.z0;
    }
    set yAxisAnnotationFormatLabel(v) {
        this.i.z0 = v;
    }
    /**
     * Gets or sets the color to use for the x axis annotation text. Leave unset for an automatic value.
    */
    get xAxisAnnotationTextColor() {
        return brushToString(this.i.acy);
    }
    set xAxisAnnotationTextColor(v) {
        this.i.acy = stringToBrush(v);
    }
    /**
     * Gets or sets the color to use for the x axis annotation backing. Leave unset for an automatic value.
    */
    get xAxisAnnotationBackground() {
        return brushToString(this.i.acw);
    }
    set xAxisAnnotationBackground(v) {
        this.i.acw = stringToBrush(v);
    }
    /**
     * Gets or sets the corner radius to use for the x axis annotation backing. Leave unset for an automatic value.
    */
    get xAxisAnnotationBackgroundCornerRadius() {
        return this.i.aad;
    }
    set xAxisAnnotationBackgroundCornerRadius(v) {
        this.i.aad = +v;
    }
    /**
     * Gets or sets the precision to use displaying values for interpolated crosshair positions.
    */
    get xAxisAnnotationInterpolatedValuePrecision() {
        return this.i.aap;
    }
    set xAxisAnnotationInterpolatedValuePrecision(v) {
        this.i.aap = +v;
    }
    /**
     * Gets or sets the color to use for the x axis annotation outline. Leave unset for an automatic value.
    */
    get xAxisAnnotationOutline() {
        return brushToString(this.i.acx);
    }
    set xAxisAnnotationOutline(v) {
        this.i.acx = stringToBrush(v);
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingLeft() {
        return this.i.aaf;
    }
    set xAxisAnnotationPaddingLeft(v) {
        this.i.aaf = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingTop() {
        return this.i.aah;
    }
    set xAxisAnnotationPaddingTop(v) {
        this.i.aah = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingRight() {
        return this.i.aag;
    }
    set xAxisAnnotationPaddingRight(v) {
        this.i.aag = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get xAxisAnnotationPaddingBottom() {
        return this.i.aae;
    }
    set xAxisAnnotationPaddingBottom(v) {
        this.i.aae = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingLeft() {
        return this.i.aal;
    }
    set yAxisAnnotationPaddingLeft(v) {
        this.i.aal = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingTop() {
        return this.i.aan;
    }
    set yAxisAnnotationPaddingTop(v) {
        this.i.aan = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingRight() {
        return this.i.aam;
    }
    set yAxisAnnotationPaddingRight(v) {
        this.i.aam = +v;
    }
    /**
     * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
    */
    get yAxisAnnotationPaddingBottom() {
        return this.i.aak;
    }
    set yAxisAnnotationPaddingBottom(v) {
        this.i.aak = +v;
    }
    /**
     * Gets or sets the stroke thickness for the x axis annotation backing. Leave unset for an automatic value.
    */
    get xAxisAnnotationStrokeThickness() {
        return this.i.aai;
    }
    set xAxisAnnotationStrokeThickness(v) {
        this.i.aai = +v;
    }
    /**
     * Gets or sets the color to use for the y axis annotation text. Leave unset for an automatic value.
    */
    get yAxisAnnotationTextColor() {
        return brushToString(this.i.ac1);
    }
    set yAxisAnnotationTextColor(v) {
        this.i.ac1 = stringToBrush(v);
    }
    /**
     * Gets or sets the color to use for the y axis annotation backing. Leave unset for an automatic value.
    */
    get yAxisAnnotationBackground() {
        return brushToString(this.i.acz);
    }
    set yAxisAnnotationBackground(v) {
        this.i.acz = stringToBrush(v);
    }
    /**
     * Gets or sets the corner radius to use for the y axis annotation backing. Leave unset for an automatic value.
    */
    get yAxisAnnotationBackgroundCornerRadius() {
        return this.i.aaj;
    }
    set yAxisAnnotationBackgroundCornerRadius(v) {
        this.i.aaj = +v;
    }
    /**
     * Gets or sets the precision to use displaying values for interpolated crosshair positions.
    */
    get yAxisAnnotationInterpolatedValuePrecision() {
        return this.i.aaq;
    }
    set yAxisAnnotationInterpolatedValuePrecision(v) {
        this.i.aaq = +v;
    }
    /**
     * Gets or sets the color to use for the y axis annotation outline. Leave unset for an automatic value.
    */
    get yAxisAnnotationOutline() {
        return brushToString(this.i.ac0);
    }
    set yAxisAnnotationOutline(v) {
        this.i.ac0 = stringToBrush(v);
    }
    /**
     * Gets or sets the stroke thickness for the y axis annotation backing. Leave unset for an automatic value.
    */
    get yAxisAnnotationStrokeThickness() {
        return this.i.aao;
    }
    set yAxisAnnotationStrokeThickness(v) {
        this.i.aao = +v;
    }
    /**
     * Gets or sets whether to skip unknown values when searching for series values.
    */
    get skipUnknownValues() {
        return this.i.z9;
    }
    set skipUnknownValues(v) {
        this.i.z9 = ensureBool(v);
    }
    get isSummarizationSupported() {
        return this.i.isSummarizationSupported;
    }
    findByName(name) {
        var baseResult = super.findByName(name);
        if (baseResult) {
            return baseResult;
        }
        if (this.targetSeries && this.targetSeries.name && this.targetSeries.name == name) {
            return this.targetSeries;
        }
        if (this.targetAxis && this.targetAxis.name && this.targetAxis.name == name) {
            return this.targetAxis;
        }
        return null;
    }
    _styling(container, component, parent) {
        super._styling(container, component, parent);
        this._inStyling = true;
        if (this.targetSeries && this.targetSeries._styling) {
            this.targetSeries._styling(container, component, this);
        }
        if (this.targetAxis && this.targetAxis._styling) {
            this.targetAxis._styling(container, component, this);
        }
        this._inStyling = false;
    }
    getSeriesValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.i5(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
}
