// igniteui-react-charts/esm2015/lib/igr-render-requested-event-args.js
import { RenderRequestedEventArgs as RenderRequestedEventArgs_internal } from "./RenderRequestedEventArgs";
import { ensureBool } from "igniteui-react-core";
/**
 * EventArgs class for RenderRequested events.
*
* Use the `RenderRequested` class for RenderRequested events.
*
* ```ts
* <IgrDataChart
* 		dataSource={this.data}
* 		ref={this.onChartRef}
* 		width="100%"
* 		height="100%" >
* <IgrCategoryXAxis name="xAxis" label="Country" />
* <IgrNumericYAxis name="yAxis" minimumValue={0} />
* <IgrColumnSeries name="series1"
* 		xAxisName="xAxis"
* 		yAxisName="yAxis"
* 		valueMemberPath="Coal"
* 		renderRequested= {this.OnRenderRequested}/>
*  </IgrDataChart>
* ```
*
* ```ts
* let series1 = new IgrColumnSeries({ name: "series1" });
* series1.valueMemberPath = "USA";
* series1.title = "USA";
* series1.xAxisName = "xAxis";
* series1.yAxisName = "yAxis";
* series1.renderRequested = this.OnRenderRequested;
*
* public OnRenderRequested =( s:IgrSeries, e:RenderRequestedEventArgs) => {
*
*   }
* ```
*/
export class IgrRenderRequestedEventArgs {
    createImplementation() {
        return new RenderRequestedEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * True if animation is requested.
    *
    * Use the `Animate` property true if the animation is requested.
    *
    * ```ts
    * e.animation = true;
    * ```
    */
    get animate() {
        return this.i.animate;
    }
    set animate(v) {
        this.i.animate = ensureBool(v);
    }
}
