// igniteui-react-charts/esm2015/lib/igr-series.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as React from 'react';
import { Visibility_$type } from "igniteui-react-core";
import { TransitionInSpeedType_$type } from './TransitionInSpeedType';
import { SeriesHitTestMode_$type } from './SeriesHitTestMode';
import { fromDoubleCollection, toDoubleCollection, brushToString, stringToBrush, toPoint, fromPoint, toRect, fromRect, stringToColor, colorToString, ensureEnum, ensureBool, isValidProp, getModifiedProps, toSpinal, initializePropertiesFromCss, NamePatcher, arrayFindByName } from "igniteui-react-core";
import { IgrRenderRequestedEventArgs } from './igr-render-requested-event-args';
import { delegateCombine, delegateRemove, TypeRegistrar } from "igniteui-react-core";
import { SeriesVisibleRangeMode_$type } from './SeriesVisibleRangeMode';
import { SeriesHighlightingMode_$type } from './SeriesHighlightingMode';
import { SeriesOutlineMode_$type } from './SeriesOutlineMode';
import { LegendItemBadgeMode_$type } from "igniteui-react-core";
import { LegendItemBadgeShape_$type } from "igniteui-react-core";
import { PenLineJoin_$type } from "igniteui-react-core";
import { PenLineCap_$type } from "igniteui-react-core";
import { IgrSeriesLayerCollection } from './igr-series-layer-collection';
import { SeriesLayerCollection as SeriesLayerCollection_internal } from './SeriesLayerCollection';
import { SeriesLayer } from './SeriesLayer';
import { SyncableObservableCollection$1 } from "igniteui-react-core";
import { TransitionOutSpeedType_$type } from './TransitionOutSpeedType';
import { SeriesHighlightedValuesDisplayMode_$type } from "igniteui-react-core";
//import { DataLegendSeriesContext, DataLegendSeriesContext_$type } from 'igniteui-core/DataLegendSeriesContext';
//import { DataLegendSeriesInfo, DataLegendSeriesInfo_$type } from 'igniteui-core/DataLegendSeriesInfo';
import { IgrSeriesLayerPropertyOverlayCollection } from './igr-series-layer-property-overlay-collection';
import { SeriesLayerPropertyOverlayCollection as SeriesLayerPropertyOverlayCollection_internal } from './SeriesLayerPropertyOverlayCollection';
import { SeriesLayerPropertyOverlay } from './SeriesLayerPropertyOverlay';
import { SeriesSelectionMode_$type } from './SeriesSelectionMode';
import { IgrTransitionOutCompletedEventArgs } from './igr-transition-out-completed-event-args';
/**
 * Represents the base class for all IgxDataChartComponent series.
*/
export class IgrSeries extends React.Component {
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    createImplementation() {
        return null;
    }
    constructor(props) {
        super(props);
        //set opacity(value: number) {
        //    this.i.opacity = +value;
        //}
        //get opacity(): number {
        //    return this.i.opacity;
        //}
        //set visibility(v: Visibility) {
        //    this.i.visibility = ensureEnum<Visibility>(Visibility_$type, v);
        //}
        //get visibility(): Visibility {
        //    return this.i.visibility;
        //}
        this._chartLevelData = null;
        this._dataSource = null;
        this._chartLevelHighlightedData = null;
        this._highlightedDataSource = null;
        this._tooltipTemplate = null;
        this._tooltipContent = null;
        this._tooltipContainerTemplate = null;
        this._renderer = null;
        this._highlightedValuesExtraPropertyOverlays = null;
        this._layers = null;
        this._actualLayers = null;
        this.__p = null;
        this._hasUserValues = new Set();
        this._stylingContainer = null;
        this._stylingParent = null;
        this._inStyling = false;
        this._transitionOutCompleted = null;
        this._transitionOutCompleted_wrapped = null;
        this._renderRequested = null;
        this._renderRequested_wrapped = null;
        if (this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
        }
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    componentDidMount() {
        for (const p of Object.keys(this.props)) {
            if (isValidProp(this, p)) {
                {
                    this[p] = this.props[p];
                }
            }
        }
    }
    get seriesInternal() {
        return this.i;
    }
    shouldComponentUpdate(nextProps, nextState) {
        const mod = getModifiedProps(this.props, nextProps);
        for (const p of Object.keys(mod)) {
            if (isValidProp(this, p)) {
                this[p] = mod[p];
            }
        }
        return true;
    }
    render() {
        return null;
    }
    set name(value) {
        this.i.name = value;
    }
    get name() {
        return this.i.name;
    }
    provideData(data) {
        this._chartLevelData = data;
        this.updateDataSource();
    }
    onUpdateDataSource(dataSource) {
        return dataSource;
    }
    updateDataSource() {
        if (this._dataSource == null) {
            this.i.itemsSource = this.onUpdateDataSource(this._chartLevelData);
        }
        else {
            this.i.itemsSource = this.onUpdateDataSource(this._dataSource);
        }
    }
    set dataSource(value) {
        this._dataSource = value;
        this.updateDataSource();
        //console.log("setting axis data source: " + value)
    }
    get dataSource() {
        if (this._dataSource != null) {
            return this._dataSource;
        }
        return this.i.itemsSource;
    }
    provideHighlightedData(data) {
        this._chartLevelHighlightedData = data;
        this.updatehighlightedDataSource();
    }
    onUpdatehighlightedDataSource(highlightedDataSource) {
        return highlightedDataSource;
    }
    updatehighlightedDataSource() {
        if (this._highlightedDataSource == null) {
            this.i.highlightedItemsSource = this.onUpdatehighlightedDataSource(this._chartLevelHighlightedData);
        }
        else {
            this.i.highlightedItemsSource = this.onUpdatehighlightedDataSource(this._highlightedDataSource);
        }
    }
    set highlightedDataSource(value) {
        this._highlightedDataSource = value;
        this.updatehighlightedDataSource();
        //console.log("setting axis data source: " + value)
    }
    get highlightedDataSource() {
        if (this._highlightedDataSource != null) {
            return this._highlightedDataSource;
        }
        return this.i.highlightedItemsSource;
    }
    bindAxes(axes) {
    }
    set tooltipTemplate(value) {
        this._tooltipTemplate = value;
        if (value == null) {
            if (this._tooltipContent !== null) {
                //this._tooltipContent.destroy();
                this._tooltipContent = null;
            }
        }
        if (this._tooltipContent != null) {
            this._tooltipContent.template = this._tooltipTemplate;
        }
        else {
            if (this.owner != null) {
                this.owner._ensureTooltipCreated(this);
            }
        }
    }
    get tooltipTemplate() {
        return this._tooltipTemplate;
    }
    set tooltipContainerTemplate(value) {
        this._tooltipContainerTemplate = value;
        if (this._tooltipContent != null) {
            this._tooltipContent.instance.containerTemplate = this._tooltipContainerTemplate;
        }
    }
    get tooltipContainerTemplate() {
        return this._tooltipContainerTemplate;
    }
    _ensureTooltipCreated(createTooltip, createWrapper) {
        if (this._tooltipTemplate == null) {
            this.i.toolTip = null;
        }
        if (this._tooltipContent == null && this._tooltipTemplate != null) {
            let tooltip = createTooltip();
            if (tooltip == null) {
                return;
            }
            this._tooltipContent = tooltip;
            tooltip.tooltipTemplate = this._tooltipTemplate;
            this.i.toolTip = createWrapper(tooltip);
        }
    }
    _ensureTooltipDestroyed() {
        if (this._tooltipContent !== null) {
            //this._tooltipContent.destroy();
            this._tooltipContent = null;
        }
    }
    _provideRenderer(renderer) {
        this._renderer = renderer;
    }
    static _createFromInternal(internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        let name = internal.$type.name;
        let externalName = "Igr" + name;
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    }
    /**
     * Gets or sets the legend used for the current series.
    */
    get legend() {
        if (this.i.legend != null)
            return this.i.legend.externalObject;
    }
    set legend(v) {
        if (v != undefined && v != null)
            this.i.legend = v.i;
    }
    /**
 * Gets unique key used to identify the series in data legend
*/
    get dataLegendKey() {
        return this.i.dataLegendKey;
    }
    /**
     * Gets or sets a name used for grouping multiple series in the Data Legend
     * If not set, series will be grouped by series family, e.g. Category, Range, Financial, Scatter
    */
    get dataLegendGroup() {
        return this.i.l3;
    }
    set dataLegendGroup(v) {
        this.i.l3 = v;
    }
    /**
     * Gets or sets a name used for grouping highlighted series in the Data Legend
     * If not set, series will default to parent series' DataLegendGroup, e.g. Category, Range, Financial, Scatter
    */
    get highlightedValuesDataLegendGroup() {
        return this.i.mj;
    }
    set highlightedValuesDataLegendGroup(v) {
        this.i.mj = v;
    }
    /**
     * Gets whether the series is an annotation layer.
    *
    * ```ts
    * let isAnnotationLayer: boolean = this.series.isAnnotationLayer;
    * ```
    */
    get isAnnotationLayer() {
        return this.i.es;
    }
    /**
     * Gets whether the series is an crosshair annotation layer.
    */
    get isAnnotationCrosshairLayer() {
        return this.i.ep;
    }
    /**
     * Gets whether the series is an callout annotation layer.
    */
    get isAnnotationCalloutLayer() {
        return this.i.eo;
    }
    /**
     * Gets whether the series is an annotation layer displayed only when hovering over the chart.
     * Crosshair Layer, Category Highlight Layer, Category Tooltip Layer, Item Tooltip Layer
    */
    get isAnnotationHoverLayer() {
        return this.i.er;
    }
    /**
     * Gets whether the series is final value annotation layer.
    */
    get isAnnotationFinalValue() {
        return this.i.eq;
    }
    /**
     * Gets whether the series is a value annotation layer.
    */
    get isAnnotationValueLayer() {
        return this.i.et;
    }
    /**
     * Gets whether the current series is oriented vertically.
    *
    * ```ts
    * let isVertical: boolean = this.series.isVertical;
    * ```
    */
    get isVertical() {
        return this.i.isVertical;
    }
    /**
     * Gets whether the current series is a stack fragment.
    *
    * ```ts
    * let isFragment: boolean = this.series.isFragment;
    * ```
    */
    get isFragment() {
        return this.i.fb;
    }
    /**
     * Gets whether the current series shows an area or line shape.
    *
    * ```ts
    * let isAreaorLine: boolean = this.series.isAreaorLine;
    * ```
    */
    get isAreaOrLine() {
        return this.i.ev;
    }
    /**
     * Gets whether the current series shows an area shape.
    */
    get isArea() {
        return this.i.eu;
    }
    /**
     * Gets whether the current series shows an area shape.
    */
    get hasValueAxis() {
        return this.i.ek;
    }
    /**
     * Gets whether the current series shows an area shape.
    */
    get isValueAxisInverted() {
        return this.i.f5;
    }
    /**
     * Gets whether the current series shows only line shapes.
    */
    get isLineOnly() {
        return this.i.fk;
    }
    /**
     * Gets whether the current series shows a line contour shape.
    */
    get isLineContour() {
        return this.i.fj;
    }
    /**
     * Gets whether the current series shows a spline shape.
    */
    get isSpline() {
        return this.i.f0;
    }
    /**
     * Gets whether the current series shows a column shape.
    */
    get isColumn() {
        return this.i.ez;
    }
    /**
     * Gets whether the current series shows a horizontal column shape.
    */
    get isBar() {
        return this.i.ex;
    }
    /**
     * Gets whether the current series shows a waterfall column shape.
    */
    get isWaterfall() {
        return this.i.f6;
    }
    /**
     * Gets whether the current series shows a polyline shape.
    */
    get isPolyline() {
        return this.i.fr;
    }
    /**
     * Gets whether the current series shows a polygon shape.
    */
    get isPolygon() {
        return this.i.fq;
    }
    /**
     * Gets whether the current series shows pixels.
    */
    get isPixel() {
        return this.i.fo;
    }
    /**
     * Gets whether the current series shows a pie shape.
    */
    get isPie() {
        return this.i.fn;
    }
    /**
     * Gets whether the current series shows a tile imagery.
    */
    get isTile() {
        return this.i.f3;
    }
    /**
     * Gets whether the current series shows step shapes.
    */
    get isStep() {
        return this.i.f1;
    }
    /**
     * Gets whether the current series shows interactive shape controls.
    */
    get isShapeControl() {
        return this.i.fy;
    }
    /**
     * Gets whether the current series is a polar type series.
    *
    * ```ts
    * let isPolar: boolean = this.series.isPolar;
    * ```
    */
    get isPolar() {
        return this.i.fp;
    }
    /**
     * Gets whether the current series is a scatter type series.
    *
    * ```ts
    * let isScatter: boolean = this.series.isScatter;
    * ```
    */
    get isScatter() {
        return this.i.fw;
    }
    /**
     * Gets whether the current series is a radial type series.
    *
    * ```ts
    * let isRadial: boolean = this.series.isRadial;
    * ```
    */
    get isRadial() {
        return this.i.ft;
    }
    /**
     * Gets whether the current series is a stacked type series.
    *
    * ```ts
    * let isStacked: boolean = this.series.isStacked;
    * ```
    */
    get isStacked() {
        return this.i.isStacked;
    }
    /**
     * Gets whether the current series is a financial type series.
    *
    * ```ts
    * let isFinancial: boolean = this.series.isFinancial;
    * ```
    */
    get isFinancial() {
        return this.i.e6;
    }
    /**
     * Gets whether the current series is a range type series.
    *
    * ```ts
    * let isRange: boolean = this.series.isRange;
    * ```
    */
    get isRange() {
        return this.i.fu;
    }
    /**
     * Gets whether the current series is a category type series.
    *
    * ```ts
    * let isCategory: boolean = this.series.isCategory;
    * ```
    */
    get isCategory() {
        return this.i.ey;
    }
    /**
     * Gets whether the current series is a geographic series.
    *
    * ```ts
    * let isGeographic: boolean = this.series.isGeographic;
    * ```
    */
    get isGeographic() {
        return this.i.fc;
    }
    /**
     * Gets whether the current series is a shape type series.
    */
    get isShape() {
        return this.i.fx;
    }
    /**
     * Gets whether the series is a value overlay.
    */
    get isValueOverlay() {
        return this.i.isValueOverlay;
    }
    /**
     * Gets whether the series is a highlight overlay.
    */
    get isHighlightOverlay() {
        return this.i.ff;
    }
    /**
     * Gets whether the series supports summarization in Data Legend and Data Tooltip.
    */
    get isSummarizationSupported() {
        return this.i.isSummarizationSupported;
    }
    get isLayer() {
        return this.i.isLayer;
    }
    /**
     * Gets whether the series has only marker as visuals
    */
    get hasOnlyMarkers() {
        return this.i.ei;
    }
    /**
     * Gets whether the series has financial OHLC/candlestick visuals
    */
    get isFinancialSeries() {
        return this.i.e9;
    }
    /**
     * Gets whether the series is financial overlay
    */
    get isFinancialOverlay() {
        return this.i.e8;
    }
    /**
     * Gets whether the series is financial indicator
    */
    get isFinancialIndicator() {
        return this.i.e7;
    }
    /**
     * Gets whether the series is financial waterfall
    */
    get isFinancialWaterfall() {
        return this.i.fa;
    }
    /**
     * Gets whether the default crosshair behavior should be disabled if this series is present.
    *
    * `IsDefaultCrosshairBehaviorDisabled` defaults to true for annotation layers. When not using annotaion layers, you can show crosshairs by setting the chart's `CrosshairVisibility` to true.
    *
    * ```ts
    * let isDefaultCrosshairBehaviorDisabled: boolean = this.series.isDefaultCrosshairBehaviorDisabled;
    * ```
    */
    get isDefaultCrosshairBehaviorDisabled() {
        return this.i.e1;
    }
    /**
     * Gets whether the default tooltip behavior should be disabled if this series is present.
    *
    * ```ts
    *  let isDefaultTooltipBehaviorDisabled: boolean = this.series.isDefaultTooltipBehaviorDisabled;
    * ```
    */
    get isDefaultTooltipBehaviorDisabled() {
        return this.i.e2;
    }
    /**
     * Returns whether the current series supports visual markers.
    *
    * ```ts
    * let hasMarkers: boolean = this.series.hasMarkers;
    * ```
    */
    get hasMarkers() {
        return this.i.hasMarkers;
    }
    /**
     * Returns whether the current series has a visible marker style.
    */
    get hasVisibleMarkers() {
        return this.i.el;
    }
    /**
     * Gets the current extra layers for the series.
    */
    get highlightedValuesExtraPropertyOverlays() {
        if (this._highlightedValuesExtraPropertyOverlays === null) {
            let coll = new IgrSeriesLayerPropertyOverlayCollection();
            let innerColl = this.i.ce;
            if (!innerColl) {
                innerColl = new SeriesLayerPropertyOverlayCollection_internal();
            }
            this._highlightedValuesExtraPropertyOverlays = coll._fromInner(innerColl);
        }
        return this._highlightedValuesExtraPropertyOverlays;
    }
    set highlightedValuesExtraPropertyOverlays(v) {
        if (this._highlightedValuesExtraPropertyOverlays !== null) {
            this._highlightedValuesExtraPropertyOverlays._setSyncTarget(null);
            this._highlightedValuesExtraPropertyOverlays = null;
        }
        let coll = new IgrSeriesLayerPropertyOverlayCollection();
        this._highlightedValuesExtraPropertyOverlays = coll._fromOuter(v);
        let syncColl = new SyncableObservableCollection$1(SeriesLayerPropertyOverlay.$type);
        let innerColl = this.i.ce;
        if (!innerColl) {
            innerColl = new SeriesLayerPropertyOverlayCollection_internal();
        }
        syncColl._inner = innerColl;
        syncColl.clear();
        this._highlightedValuesExtraPropertyOverlays._setSyncTarget(syncColl);
    }
    /**
     * Gets if the series uses an index, which can be used to auto-assign it a color.
    *
    * ```ts
    * let isIndexed: boolean = this.series.isIndexed;
    * ```
    */
    get isIndexed() {
        return this.i.fg;
    }
    /**
     * Gets if the series should appear in any legends.
    *
    * ```ts
    * let isUsableInLegend: boolean = this.series.isUsableInLegend;
    * ```
    */
    get isUsableInLegend() {
        return this.i.isUsableInLegend;
    }
    /**
     * Gets or sets whether or not the current series will have a legend item displayed in a legend.
    *
    * The `LegendItemVisibilty` can be used to exclude only certain `Series` from the `Legend`.
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     legendItemVisibility= "collapsed" />
    *  </IgrDataChart>
    * ```
    */
    get legendItemVisibility() {
        return this.i.legendItemVisibility;
    }
    set legendItemVisibility(v) {
        this.i.legendItemVisibility = ensureEnum(Visibility_$type, v);
    }
    /**
     * Gets or sets the LegendItemBadgeTemplate property.
     * The legend item badge is created according to the LegendItemBadgeTemplate on-demand by
     * the series object itself.
    */
    get legendItemBadgeTemplate() {
        return this.i.legendItemBadgeTemplate;
    }
    set legendItemBadgeTemplate(v) {
        this.i.legendItemBadgeTemplate = v;
    }
    /**
     * Gets the actual legend item badge template used by the series.
    */
    get actualLegendItemBadgeTemplate() {
        return this.i.sw;
    }
    set actualLegendItemBadgeTemplate(v) {
        this.i.sw = v;
    }
    /**
     * Gets the effective legend badge outline for the current series.
    */
    get actualLegendItemBadgeOutline() {
        return brushToString(this.i.vc);
    }
    set actualLegendItemBadgeOutline(v) {
        this.i.vc = stringToBrush(v);
    }
    /**
     * Gets the effective legend badge fill for the current series.
    */
    get actualLegendItemBadgeBrush() {
        return brushToString(this.i.vb);
    }
    set actualLegendItemBadgeBrush(v) {
        this.i.vb = stringToBrush(v);
    }
    /**
     * Gets or sets the type of legend badge representing the current series in a legend.
     * This property will be ignored when the LegendItemTemplate/LegendItemBadgeTemplate property is set on the series
    */
    get legendItemBadgeShape() {
        return this.i.legendItemBadgeShape;
    }
    set legendItemBadgeShape(v) {
        this.i.legendItemBadgeShape = ensureEnum(LegendItemBadgeShape_$type, v);
    }
    /**
     * Gets or sets the mode of legend badge representing the current series in a legend.
     * This property will be ignored when the LegendItemTemplate/LegendItemBadgeTemplate property is set on the series
    */
    get legendItemBadgeMode() {
        return this.i.bd;
    }
    set legendItemBadgeMode(v) {
        this.i.bd = ensureEnum(LegendItemBadgeMode_$type, v);
    }
    /**
     * Gets or sets the LegendItemTemplate property.
     * The legend item control content is created according to the LegendItemTemplate on-demand by
     * the series object itself.
    */
    get legendItemTemplate() {
        return this.i.legendItemTemplate;
    }
    set legendItemTemplate(v) {
        this.i.legendItemTemplate = v;
    }
    /**
     * Gets or sets the DiscreteLegendItemTemplate property.
     * The legend item control content is created according to the DiscreteLegendItemTemplate on-demand by
     * the series object itself.
    */
    get discreteLegendItemTemplate() {
        return this.i.sx;
    }
    set discreteLegendItemTemplate(v) {
        this.i.sx = v;
    }
    /**
     * Gets the Index property.
    */
    get index() {
        return this.i.index;
    }
    set index(v) {
        this.i.index = +v;
    }
    /**
     * Gets or sets the easing function used to morph the current series.
    *
    * The `TransitioninDuration` and `TransitionEasingFunction` can be used to play animation when data is added or removed from a `Series`. To play an initial animation see `TransitionInDuration`.
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     isDropShadowEnabled=true
    *     markerType="circle"
    *    transitionEasingFunction ="cubicEase"
    *    isTransitionInEnabled= {true}  />
    *  </IgrDataChart>
    * ```
    *
    * ```ts
    * this.series.transitionEasingFunction =  EasingFunctions.cubicEase;
    * ```
    */
    get transitionEasingFunction() {
        return this.i.dd;
    }
    set transitionEasingFunction(v) {
        this.i.dd = v;
    }
    /**
     * Gets or sets the EasingFunction used to morph the current series during the initial transition.
    *
    * The `TransitionInEasingFunction` can be used with `IsTransitionInEnabled` to cofigure the animation when a new datasource is loaded.
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     isDropShadowEnabled=true
    *     markerType="circle"
    *    transitionInEasingFunction ="cubicEase"
    *    isTransitionInEnabled= {true}  />
    *  </IgrDataChart>
    * ```
    *
    * ```ts
    * this.series.transitionInEasingFunction =  EasingFunctions.cubicEase;
    * ```
    */
    get transitionInEasingFunction() {
        return this.i.de;
    }
    set transitionInEasingFunction(v) {
        this.i.de = v;
    }
    /**
     * Gets or sets the EasingFunction used to morph the current series during the initial transition.
    */
    get transitionOutEasingFunction() {
        return this.i.df;
    }
    set transitionOutEasingFunction(v) {
        this.i.df = v;
    }
    /**
     * Gets or sets the duration of the current series' morph.
    *
    * The `TransitionDuration` can be used to play animation when data is added or removed from a `Series`. To play an initial animation see `TransitionInDuration`.
    *
    * ```ts
    * this.series.transitionDuration = 500;
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     isTransitionInEnabled=true
    *     transitionInSpeedType="indexScaled"
    *     transitionDuration= {500}
    *     transitionInDuration = {500} />
    *  </IgrDataChart>
    * ```
    */
    get transitionDuration() {
        return this.i.kd;
    }
    set transitionDuration(v) {
        this.i.kd = +v;
    }
    get actualResolution() {
        return this.i.id;
    }
    set actualResolution(v) {
        this.i.id = +v;
    }
    /**
     * Gets or sets the current series object's rendering resolution.
    *
    * Setting the `Resolution` on a Series to a higher value can help with performance, but it will lower the graphical fidelity of the line. As such, it can be increased up until the fidelity is unacceptable.
    *
    * ```ts
    * this.series.resolution = 1.5;
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     resolution= {1.5} />
    *  </IgrDataChart>
    * ```
    */
    get resolution() {
        return this.i.jj;
    }
    set resolution(v) {
        this.i.jj = +v;
    }
    /**
     * Gets or sets the top margin to use when getting a visible axis range for the series.
    */
    get visibleRangeMarginTop() {
        return this.i.ju;
    }
    set visibleRangeMarginTop(v) {
        this.i.ju = +v;
    }
    /**
     * Gets or sets the bottom margin to use when getting a visible axis range for the series.
    */
    get visibleRangeMarginBottom() {
        return this.i.jr;
    }
    set visibleRangeMarginBottom(v) {
        this.i.jr = +v;
    }
    /**
     * Gets or sets the left margin to use when getting a visible axis range for the series.
    */
    get visibleRangeMarginLeft() {
        return this.i.js;
    }
    set visibleRangeMarginLeft(v) {
        this.i.js = +v;
    }
    /**
     * Gets or sets the right margin to use when getting a visible axis range for the series.
    */
    get visibleRangeMarginRight() {
        return this.i.jt;
    }
    set visibleRangeMarginRight(v) {
        this.i.jt = +v;
    }
    /**
     * Gets or sets the Title property.
     * The legend item control is created according to the Title on-demand by
     * the series object itself.
    *
    * The `Series` `Title` may be used in tooltips and legends.
    *
    * ```ts
    * this.series.title = "InStock Items";
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *    title="InStock Items" />
    *  </IgrDataChart>
    * ```
    */
    get title() {
        return this.i.title;
    }
    set title(v) {
        this.i.title = v;
    }
    /**
     * Gets or sets the HighlightedTitleSuffix property.
    */
    get highlightedTitleSuffix() {
        return this.i.mh;
    }
    set highlightedTitleSuffix(v) {
        this.i.mh = v;
    }
    /**
     * Gets or sets whether the highlighted values layer should have a legend item.
    */
    get highlightedLegendItemVisibility() {
        return this.i.wx;
    }
    set highlightedLegendItemVisibility(v) {
        this.i.wx = ensureEnum(Visibility_$type, v);
    }
    /**
     * Gets or sets the brush to use for the series.
    *
    * The `Brush` along with the `Outline` and `Negative Brush` can be used to affect the visuals of the `Series`.
    *
    * ```ts
    * this.series.brush = "red";
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     brush="Gray"
    *     outline="Black" />
    *  </IgrDataChart>
    * ```
    */
    get brush() {
        return brushToString(this.i.vf);
    }
    set brush(v) {
        this.i.vf = stringToBrush(v);
    }
    /**
     * Gets the effective brush for the current series object.
    *
    * ```ts
    * let actualBrush: string = series.actualBrush;
    * ```
    */
    get actualBrush() {
        return brushToString(this.i.u9);
    }
    set actualBrush(v) {
        this.i.u9 = stringToBrush(v);
    }
    /**
     * Gets or sets the selection brush to use for the series.
    */
    get selectionBrush() {
        return brushToString(this.i.vz);
    }
    set selectionBrush(v) {
        this.i.vz = stringToBrush(v);
    }
    /**
     * Gets or sets the Focus brush to use for the series.
    */
    get focusBrush() {
        return brushToString(this.i.vk);
    }
    set focusBrush(v) {
        this.i.vk = stringToBrush(v);
    }
    /**
     * Gets the effective emphasis brush for the current series object.
    */
    get actualSelectionBrush() {
        return brushToString(this.i.ve);
    }
    set actualSelectionBrush(v) {
        this.i.ve = stringToBrush(v);
    }
    /**
     * Gets the effective emphasis brush for the current series object.
    */
    get actualFocusBrush() {
        return brushToString(this.i.va);
    }
    set actualFocusBrush(v) {
        this.i.va = stringToBrush(v);
    }
    /**
     * Gets the effective brush for the current series object with opacity removed so its contrasty for use as a font color.
    */
    get safeActualBrush() {
        return brushToString(this.i.safeActualBrush);
    }
    set safeActualBrush(v) {
        this.i.safeActualBrush = stringToBrush(v);
    }
    /**
     * Gets or sets the brush to use for the outline of the series.
     * Some series types, such as LineSeries, do not display outlines.  Therefore, this property does not affect some charts.
    *
    * The `Outline` along with the `Brush` and `NegativeBrush` can be used to affect the visuals of the `Series`.
    *
    * ```ts
    * this.series.brush = "red";
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     brush="Gray"
    *     outline="Black" />
    *  </IgrDataChart>
    * ```
    */
    get outline() {
        return brushToString(this.i.vv);
    }
    set outline(v) {
        this.i.vv = stringToBrush(v);
    }
    /**
     * Gets the effective outline for the current series object.
    *
    * ```ts
    * let actualOutline: string = series.ActualOutline;
    * ```
    */
    get actualOutline() {
        return brushToString(this.i.vd);
    }
    set actualOutline(v) {
        this.i.vd = stringToBrush(v);
    }
    /**
     * Gets or sets the brush that specifies current series object's line join style.
    */
    get lineJoin() {
        return this.i.v5;
    }
    set lineJoin(v) {
        this.i.v5 = ensureEnum(PenLineJoin_$type, v);
    }
    /**
     * Gets or sets whether and how to display highlighted values for the series. Note, this is distinct from the highlighting feature that indicates what is closest or under the mouse.
    */
    get highlightedValuesDisplayMode() {
        return this.i.bv;
    }
    set highlightedValuesDisplayMode(v) {
        this.i.bv = ensureEnum(SeriesHighlightedValuesDisplayMode_$type, v);
    }
    /**
     * Gets or sets whether and how to display highlighted values for the series. Note, this is distinct from the highlighting feature that indicates what is closest or under the mouse.
    */
    get shouldAnimateOnDataSourceSwap() {
        return this.i.gg;
    }
    set shouldAnimateOnDataSourceSwap(v) {
        this.i.gg = ensureBool(v);
    }
    /**
     * Gets or sets the width of the current series object's line thickness.
    *
    * Depending on the `Series` type, this can be the main brush used, or just the outline. For example, when using a `LineSeries` it will affect the thickness of the lines drawn, whereas when using a `ColumnSeries` it will affect the outer border thickness of the columns.
    *
    * ```ts
    * this.series.thickness=5;
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     thickness= {5} />
    *  </IgrDataChart>
    * ```
    */
    get thickness() {
        return this.i.jp;
    }
    set thickness(v) {
        this.i.jp = +v;
    }
    /**
     * Gets the effective Thickness for the current series object.
    */
    get actualThickness() {
        return this.i.ie;
    }
    set actualThickness(v) {
        this.i.ie = +v;
    }
    /**
     * Gets or sets a collection of Double values that indicate the pattern of dashes and gaps that
     * is used to outline the current series object.
    */
    get dashArray() {
        return fromDoubleCollection(this.i.v2);
    }
    set dashArray(v) {
        this.i.v2 = toDoubleCollection(v);
    }
    /**
     * Gets actual highlighting mode
    */
    get actualHighlightingMode() {
        return this.i.bx;
    }
    set actualHighlightingMode(v) {
        this.i.bx = ensureEnum(SeriesHighlightingMode_$type, v);
    }
    /**
     * Gets actual Selection mode
    */
    get actualSelectionMode() {
        return this.i.cp;
    }
    set actualSelectionMode(v) {
        this.i.cp = ensureEnum(SeriesSelectionMode_$type, v);
    }
    /**
     * Gets actual Selection mode
    */
    get actualFocusMode() {
        return this.i.co;
    }
    set actualFocusMode(v) {
        this.i.co = ensureEnum(SeriesSelectionMode_$type, v);
    }
    /**
     * Gets actual HighlightedValues fade opacity
    */
    get actualHighlightedValuesFadeOpacity() {
        return this.i.ia;
    }
    set actualHighlightedValuesFadeOpacity(v) {
        this.i.ia = +v;
    }
    /**
     * Gets or sets the target opacity to fade to for fade style HighlightedValues.
    */
    get highlightedValuesFadeOpacity() {
        return this.i.je;
    }
    set highlightedValuesFadeOpacity(v) {
        this.i.je = +v;
    }
    /**
     * Gets actual highlighting fade opacity
    */
    get actualHighlightingFadeOpacity() {
        return this.i.ib;
    }
    set actualHighlightingFadeOpacity(v) {
        this.i.ib = +v;
    }
    /**
     * Gets or sets the target opacity to fade to for fade style highlighting.
    */
    get highlightingFadeOpacity() {
        return this.i.jf;
    }
    set highlightingFadeOpacity(v) {
        this.i.jf = +v;
    }
    /**
     * Gets or sets whether the ActualLegend is FinancialLegend or normal Legend.
    */
    get isActualLegendFinancial() {
        return this.i.en;
    }
    set isActualLegendFinancial(v) {
        this.i.en = ensureBool(v);
    }
    /**
     * Gets or sets whether the component level highlight mode is ignored.
    */
    get isComponentHighlightingModeIgnored() {
        return this.i.e0;
    }
    set isComponentHighlightingModeIgnored(v) {
        this.i.e0 = ensureBool(v);
    }
    /**
     * Gets or sets whether highlighting should be enabled for this series, if this type of series supports highlighting.
    *
    * ```ts
    * this.series.isHighlightingEnabled = true;
    * ```
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     isHighlightingEnabled="true" />
    *  </IgrDataChart>
    * ```
    */
    get isHighlightingEnabled() {
        return this.i.fd;
    }
    set isHighlightingEnabled(v) {
        this.i.fd = ensureBool(v);
    }
    /**
     * Gets or sets whether the series should use individual palette colors for each item.
    */
    get useItemWiseColors() {
        return this.i.g9;
    }
    set useItemWiseColors(v) {
        this.i.g9 = ensureBool(v);
    }
    get isColoredItemwise() {
        return this.i.isColoredItemwise;
    }
    /**
     * Gets or sets whether the opacity should be automatically shifted for the safe actual brush.
    */
    get shouldShiftOpacityForSafeActualBrush() {
        return this.i.gl;
    }
    set shouldShiftOpacityForSafeActualBrush(v) {
        this.i.gl = ensureBool(v);
    }
    /**
     * Gets or sets whether the opacity should be automatically shifted for the safe actual brush.
    */
    get shouldRemoveHighlightedDataOnLayerHidden() {
        return this.i.gj;
    }
    set shouldRemoveHighlightedDataOnLayerHidden(v) {
        this.i.gj = ensureBool(v);
    }
    /**
     * Gets or sets whether this series should suppress it's auto callouts
    */
    get shouldHideAutoCallouts() {
        return this.i.gi;
    }
    set shouldHideAutoCallouts(v) {
        this.i.gi = ensureBool(v);
    }
    /**
     * Gets or sets whether drop shadow should be enabled for this series.
    *
    * `IsDropShadowEnabled` is used to decide whether drop shadow should be enabled for this series.
    *
    * ```ts
    * this.series.isDropShadowEnabled = true;
    * ```
    *
    * ```ts
    * <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     isDropShadowEnabled="true" />
    *  </IgrDataChart>
    * ```
    */
    get isDropShadowEnabled() {
        return this.i.e4;
    }
    set isDropShadowEnabled(v) {
        this.i.e4 = ensureBool(v);
    }
    /**
     * Gets or sets the shadow blur.
     * This property is ignored when
     * Series.UseSingleShadow is set to true.
    *
    * `ShadowBlur` can be set in conjection with `IsDropShadowEnabled` to further define the drop shadow effect.
    *
    * ```ts
    * this.series.shadowBlur = 7;
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     shadowBlur={7}
    *     shadowColor="blue" />
    *  </IgrDataChart>
    * ```
    */
    get shadowBlur() {
        return this.i.jm;
    }
    set shadowBlur(v) {
        this.i.jm = +v;
    }
    /**
     * Gets or sets the drop shadow color.
    *
    * `ShadowColor` can be set in conjection with `IsDropShadowEnabled` to further define the drop shadow effect.
    *
    * ```ts
    * this.series.shadowColor = "blue";
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     shadowBlur="7"
    *     shadowColor="blue" />
    *  </IgrDataChart>
    * ```
    */
    get shadowColor() {
        return colorToString(this.i.v1);
    }
    set shadowColor(v) {
        this.i.v1 = stringToColor(v);
    }
    /**
     * Gets or sets whether drop shadow is applied to the whole series visual or to each of the individual shapes forming the series.
     * When this property is set to true, no
     * Series.ShadowBlur is applied.
    *
    * ```ts
    * this.series.useSingleShadow = true;
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     isDropShadowEnabled=true
    *     markerType="circle"
    *     useSingleShadow= {true} />
    *  </IgrDataChart>
    * ```
    */
    get useSingleShadow() {
        return this.i.hc;
    }
    set useSingleShadow(v) {
        this.i.hc = ensureBool(v);
    }
    /**
     * Gets or sets the drop shadow x-offset.
    *
    * `ShadowOffsetX` can be set in conjection with `IsDropShadowEnabled` to further define the drop shadow effect.
    *
    * ```ts
    * this.series.shadowOffsetX = 10;
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     shadowOffsetX= {10}
    *     shadowOffsetY={10] />
    *  </IgrDataChart>
    * ```
    */
    get shadowOffsetX() {
        return this.i.jn;
    }
    set shadowOffsetX(v) {
        this.i.jn = +v;
    }
    /**
     * Gets or sets the drop shadow y-offset.
    *
    * `ShadowOffsetY` can be set in conjection with `IsDropShadowEnabled` to further define the drop shadow effect.
    *
    * ```ts
    * this.series.shadowOffsetY = 10;
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     shadowOffsetX= {10}
    *     shadowOffsetY= {10} />
    *  </IgrDataChart>
    * ```
    */
    get shadowOffsetY() {
        return this.i.jo;
    }
    set shadowOffsetY(v) {
        this.i.jo = +v;
    }
    /**
     * Gets or sets the opacity applied to the area fill visual.
     * This property only applies to series that have area visual.
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *    actualAreaFillOpacity= {1} />
    *  </IgrDataChart>
    * ```
    */
    get areaFillOpacity() {
        return this.i.ik;
    }
    set areaFillOpacity(v) {
        this.i.ik = +v;
    }
    /**
     * Gets the actual opacity applied to the area fill visual.
    *
    * ```ts
    * let actualFillOpacity: number = series.actualAreaFillOpacity;
    * ```
    */
    get actualAreaFillOpacity() {
        return this.i.h9;
    }
    set actualAreaFillOpacity(v) {
        this.i.h9 = +v;
    }
    /**
     * Gets or sets the opacity applied to the fill of the markers.
     * This property only applies to series that have area visual.
    */
    get markerFillOpacity() {
        return this.i.jg;
    }
    set markerFillOpacity(v) {
        this.i.jg = +v;
    }
    /**
     * Gets the actual opacity applied to the fill of the markers.
    */
    get actualMarkerFillOpacity() {
        return this.i.ic;
    }
    set actualMarkerFillOpacity(v) {
        this.i.ic = +v;
    }
    /**
     * Gets if the default tooltip has been selected for display.
    *
    * To use the default tooltip set `ShowDefaultTooltip` to true.
    *
    * ```ts
    * let isDefaultToolTipSelected: boolean = this.series.isDefaultToolTipSelected;
    * ```
    */
    get isDefaultToolTipSelected() {
        return this.i.e3;
    }
    set isDefaultToolTipSelected(v) {
        this.i.e3 = ensureBool(v);
    }
    /**
     * Gets or sets whether default tooltip will be shown.
    *
    * The default tooltips display all the information relevant to the particular series item (series title, data values, axis values etc.) and are styled to match the series' style. When using default tooltips, you should also set the series `Title`.
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     showDefaultTooltip ="true"
    *     title="InStock Items" />
    *  </IgrDataChart>
    * ```
    *
    * ```ts
    * this.series.showDefaultTooltip ="true";
    * ```
    */
    get showDefaultTooltip() {
        return this.i.gp;
    }
    set showDefaultTooltip(v) {
        this.i.gp = ensureBool(v);
    }
    /**
     * Gets or sets the visible range mode to use.
    */
    get visibleRangeMode() {
        return this.i.cz;
    }
    set visibleRangeMode(v) {
        this.i.cz = ensureEnum(SeriesVisibleRangeMode_$type, v);
    }
    /**
     * Gets or sets the outline mode to use for the series.
    */
    get outlineMode() {
        return this.i.ci;
    }
    set outlineMode(v) {
        this.i.ci = ensureEnum(SeriesOutlineMode_$type, v);
    }
    /**
     * Gets or sets the duration of the current series' transition in morph.
    *
    * The `TransitionInDuration` can be used with `IsTransitionInEnabled` to cofigure the animation when a new datasource is loaded.
    *
    * ```ts
    * this.series.transitionDuration = 500;
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     isTransitionInEnabled= {true}
    *     transitionInSpeedType="indexScaled"
    *     transitionDuration={500}
    *     transitionInDuration= {500} />
    *  </IgrDataChart>
    * ```
    */
    get transitionInDuration() {
        return this.i.ke;
    }
    set transitionInDuration(v) {
        this.i.ke = +v;
    }
    /**
     * Gets or sets the duration of the current series' transition out morph.
    */
    get transitionOutDuration() {
        return this.i.kf;
    }
    set transitionOutDuration(v) {
        this.i.kf = +v;
    }
    /**
     * Gets or sets the duration of the current series' transition in morph.
    *
    * The `TransitionInSpeedType` can be used with `IsTransitionInEnabled` to cofigure the animation when a new datasource is loaded.
    *
    * ```ts
    * this.series.transitionInSpeedType = TransitionInSpeedType.IndexScaled;
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     isTransitionInEnabled=true
    *     transitionInSpeedType="indexScaled"
    *     transitionDuration=500
    *     transitionInDuration=500 />
    *  </IgrDataChart>
    * ```
    */
    get transitionInSpeedType() {
        return this.i.c5;
    }
    set transitionInSpeedType(v) {
        this.i.c5 = ensureEnum(TransitionInSpeedType_$type, v);
    }
    /**
     * Gets or sets the duration of the current series' transition out morph.
    */
    get transitionOutSpeedType() {
        return this.i.c8;
    }
    set transitionOutSpeedType(v) {
        this.i.c8 = ensureEnum(TransitionOutSpeedType_$type, v);
    }
    /**
     * Gets or sets the style of the starting point of any lines or polylines representing this series.
     * Not every series type has a line at which it would be appropriate to display a start cap, so this property does not affect every series type.  LineSeries, for example, is affected by StartCap, but ColumnSeries is not.
    */
    get lineCap() {
        return this.i.v4;
    }
    set lineCap(v) {
        this.i.v4 = ensureEnum(PenLineCap_$type, v);
    }
    /**
     * Gets or sets the AutoCalloutLabel format string to use for the AutoCalloutLabel.
    */
    get autoCalloutLabelFormat() {
        return this.i.lu;
    }
    set autoCalloutLabelFormat(v) {
        this.i.lu = v;
    }
    /**
     * Gets or sets the format specifiers to use with the AutoCalloutLabelFormat string.
    */
    get autoCalloutLabelFormatSpecifiers() {
        return this.i.ag;
    }
    set autoCalloutLabelFormatSpecifiers(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.ag = v;
    }
    /**
     * Gets or sets the AutoCalloutLabel format string to use for the AutoCalloutLabel.
    */
    get autoCalloutValueLabelFormat() {
        return this.i.lx;
    }
    set autoCalloutValueLabelFormat(v) {
        this.i.lx = v;
    }
    /**
     * Gets or sets the format specifiers to use with the AutoCalloutValueLabelFormat string.
    */
    get autoCalloutValueLabelFormatSpecifiers() {
        return this.i.ah;
    }
    set autoCalloutValueLabelFormatSpecifiers(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.ah = v;
    }
    /**
     * Gets or sets the whether the chart reacts to mouse move events.
    */
    get mouseOverEnabled() {
        return this.i.f8;
    }
    set mouseOverEnabled(v) {
        this.i.f8 = ensureBool(v);
    }
    /**
     * Gets or sets the coercion methods to use when loading data from data sources.
     * Should be specified before setting any member paths, if being used. Setting it later
     * will not cause data to be reimported into the chart.
    *
    * &lt;-- position: content member-->
    *
    * ```ts
    * cMethods: any = {
    *     ValueCoercion: function(value:number): number {return value+2;}
    *  }
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource}
    *     width="700px"
    *     height="500px">
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrScatterPolygonSeries
    *      name="series1"
    *      shapeMemberPath="Points"
    *      title="House Floor Plan"
    *      brush="Gray"
    *      outline="Black"
    * 	 coercionMethods = {this.ValueCoercion} />
    *  </IgrDataChart>
    * ```
    */
    get coercionMethods() {
        return this.i.coercionMethods;
    }
    set coercionMethods(v) {
        this.i.coercionMethods = v;
    }
    /**
     * Gets or sets whether the series should expect that properties in its items source may be functions that need to be
     * evaluated to yield their value. This defaults to false, as there will be a subtle performance impact to using this feature.
     * This value must be set before assigning an item's source to take effect.
    *
    * `expectFunctions` is used to decide whether the series should expect that its data source members need to be called as functions to get their values.
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="MyFunction"
    *     expectFunctions={true} />
    *  </IgrDataChart>
    * ```
    *
    * ```ts
    * this.series.expectFunctions=true;
    * ```
    */
    get expectFunctions() {
        return this.i.expectFunctions;
    }
    set expectFunctions(v) {
        this.i.expectFunctions = ensureBool(v);
    }
    /**
     * Method of hit testing to use when pointing to items in this series.
    *
    * The `HitTestMode` can be set to either:
    *
    * -   `Auto`: automatically decide the appropriate hit test mode for the series.
    * -   `ColorEncoded`: use a color encoded off screen buffer for hit testing. Should always be O(1) time for determining a hit series. This increases frame render time and memory usage, however. Consider using this if hit testing time is degrading performance.
    * -   `Computational`: use a computational based approach to determine whether the series has been hit. This uses loose bounding boxes, in some cases, and can range in time complexity between O(1) and O(log n) to find a hit. This decreases frame render time compared to color encoded
    *
    * ```ts
    * this.series.hitTestMode = SeriesHitTestMode.Computational;
    * ```
    */
    get hitTestMode() {
        return this.i.b3;
    }
    set hitTestMode(v) {
        this.i.b3 = ensureEnum(SeriesHitTestMode_$type, v);
    }
    /**
     * Resolved method of hit testing to use when pointing to items in the chart.
    *
    * ```ts
    * let hitTestMode: SeriesHitTestMode = this.series.actualHitTestMode;
    * ```
    */
    get actualHitTestMode() {
        return this.i.b1;
    }
    set actualHitTestMode(v) {
        this.i.b1 = ensureEnum(SeriesHitTestMode_$type, v);
    }
    /**
     * Overridden by derived series classes to indicate when negative colors are supported or not.
    */
    get isNegativeColorSupported() {
        return this.i.fm;
    }
    /**
     * Overridden by derived series classes to indicate when markerless display is preferred or not.
    */
    get isMarkerlessDisplayPreferred() {
        return this.i.fl;
    }
    /**
     * The final value of this series.
    */
    get finalValue() {
        return this.i.im;
    }
    set finalValue(v) {
        this.i.im = +v;
    }
    /**
     * The percent change from the beginning to the end of this series.
    */
    get percentChange() {
        return this.i.ji;
    }
    set percentChange(v) {
        this.i.ji = +v;
    }
    /**
     * Gets the text that is displayed for the Open label in the tooltip.
    */
    get actualOpenLabel() {
        return this.i.lk;
    }
    /**
     * Gets the text that is displayed for the Close label in the tooltip.
    */
    get actualCloseLabel() {
        return this.i.k6;
    }
    /**
     * Gets the text that is displayed for the High label in the tooltip.
    */
    get actualHighLabel() {
        return this.i.k9;
    }
    /**
     * Gets the text that is displayed for the Low label in the tooltip.
    */
    get actualLowLabel() {
        return this.i.li;
    }
    /**
     * Gets the text that is displayed for the Volume label in the tooltip.
    */
    get actualVolumeLabel() {
        return this.i.ls;
    }
    /**
     * Gets the text that is displayed for the Value label in the tooltip.
    */
    get actualValueLabel() {
        return this.i.lr;
    }
    /**
     * Gets the text that is displayed for the Value label in the tooltip.
    */
    get actualRadiusLabel() {
        return this.i.lm;
    }
    /**
     * Gets the current extra layers for the series.
    */
    get layers() {
        if (this._layers === null) {
            let coll = new IgrSeriesLayerCollection();
            let innerColl = this.i.cb;
            if (!innerColl) {
                innerColl = new SeriesLayerCollection_internal();
            }
            this._layers = coll._fromInner(innerColl);
        }
        return this._layers;
    }
    set layers(v) {
        if (this._layers !== null) {
            this._layers._setSyncTarget(null);
            this._layers = null;
        }
        let coll = new IgrSeriesLayerCollection();
        this._layers = coll._fromOuter(v);
        let syncColl = new SyncableObservableCollection$1(SeriesLayer.$type);
        let innerColl = this.i.cb;
        if (!innerColl) {
            innerColl = new SeriesLayerCollection_internal();
        }
        syncColl._inner = innerColl;
        syncColl.clear();
        this._layers._setSyncTarget(syncColl);
    }
    /**
     * Gets the actual extra layers for the series.
    */
    get actualLayers() {
        if (this._actualLayers === null) {
            let coll = new IgrSeriesLayerCollection();
            let innerColl = this.i.actualLayers;
            if (!innerColl) {
                innerColl = new SeriesLayerCollection_internal();
            }
            this._actualLayers = coll._fromInner(innerColl);
        }
        return this._actualLayers;
    }
    set actualLayers(v) {
        if (this._actualLayers !== null) {
            this._actualLayers._setSyncTarget(null);
            this._actualLayers = null;
        }
        let coll = new IgrSeriesLayerCollection();
        this._actualLayers = coll._fromOuter(v);
        let syncColl = new SyncableObservableCollection$1(SeriesLayer.$type);
        let innerColl = this.i.actualLayers;
        if (!innerColl) {
            innerColl = new SeriesLayerCollection_internal();
        }
        syncColl._inner = innerColl;
        syncColl.clear();
        this._actualLayers._setSyncTarget(syncColl);
    }
    get opacity() {
        return this.i.opacity;
    }
    set opacity(v) {
        this.i.opacity = +v;
    }
    get visibility() {
        return this.i.visibility;
    }
    set visibility(v) {
        this.i.visibility = ensureEnum(Visibility_$type, v);
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.highlightedValuesExtraPropertyOverlays != null && this.highlightedValuesExtraPropertyOverlays.findByName && this.highlightedValuesExtraPropertyOverlays.findByName(name)) {
            return this.highlightedValuesExtraPropertyOverlays.findByName(name);
        }
        if (this.legend && this.legend.name && this.legend.name == name) {
            return this.legend;
        }
        if (this.autoCalloutLabelFormatSpecifiers != null && arrayFindByName(this.autoCalloutLabelFormatSpecifiers, name)) {
            return arrayFindByName(this.autoCalloutLabelFormatSpecifiers, name);
        }
        if (this.autoCalloutValueLabelFormatSpecifiers != null && arrayFindByName(this.autoCalloutValueLabelFormatSpecifiers, name)) {
            return arrayFindByName(this.autoCalloutValueLabelFormatSpecifiers, name);
        }
        if (this.layers != null && this.layers.findByName && this.layers.findByName(name)) {
            return this.layers.findByName(name);
        }
        if (this.actualLayers != null && this.actualLayers.findByName && this.actualLayers.findByName(name)) {
            return this.actualLayers.findByName(name);
        }
        return null;
    }
    get hasUserValues() {
        return this._hasUserValues;
    }
    __m(propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    }
    _styling(container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        let genericPrefix = "";
        let typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("Series");
        let additionalPrefixes = [];
        let prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        let b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            let basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            let parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            let parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    }
    /**
 * Gets the brush for tooltip labels.

*/
    resolveTooltipBrush() {
        let iv = this.i.ne();
        return (iv);
    }
    getItemValue(item, memberPathName) {
        let iv = this.i.kq(item, memberPathName);
        return (iv);
    }
    /**
     * Gets the value of a requested member path from the series.
    
    * @param memberPathName  * The property name of a valid member path for the series
    */
    getMemberPathValue(memberPathName) {
        let iv = this.i.md(memberPathName);
        return (iv);
    }
    /**
     * Gets the precise item index, if possible, based on the closeness to the previous or next whole integer. If the series cannot provide this information, GetExactItemIndex will return the same integer value as GetItemIndex.
    
    * @param world  * The world position for which to return the index.
    *
    * ```ts
    * let index: number = this.series.getExactItemIndex({x:.5,y:.5});
    * ```
    */
    getExactItemIndex(world) {
        let iv = this.i.ir(toPoint(world));
        return (iv);
    }
    /**
     * Gets the item index associated with the specified world position
    
    *
    * ```ts
    * let index: number = this.series.getItemIndex({x:.5,y:.5});
    * ```
    */
    getItemIndex(world) {
        let iv = this.i.j3(toPoint(world));
        return (iv);
    }
    /**
     * Gets the item that is the best match for the specified world coordinates.
    
    * @param world  * The world coordinates to use.
    *
    * ```ts
    * let item: number = this.series.getItem({x:.5,y:.5});
    * ```
    */
    getItem(world) {
        let iv = this.i.kn(toPoint(world));
        return (iv);
    }
    getPreviousOrExactIndex(world, skipUnknowns) {
        let iv = this.i.ka(toPoint(world), skipUnknowns);
        return (iv);
    }
    getNextOrExactIndex(world, skipUnknowns) {
        let iv = this.i.j8(toPoint(world), skipUnknowns);
        return (iv);
    }
    getSeriesValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.i5(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.wc(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    /**
     * If possible, will return the best available value bounding box within the series that has the best value match for the world position provided.
    
    * @param world  * The world coordinate for which to get a value bounding box for
    */
    getSeriesValueBoundingBox(world) {
        let iv = this.i.wn(toPoint(world));
        return fromRect(iv);
    }
    /**
     * If possible, will return the best available value marker bounding box within the series that has the best value match for the world position provided.
    
    * @param world  * The world coordinates for which to get a value marker bounding box for
    */
    getSeriesValueMarkerBoundingBox(world) {
        let iv = this.i.wp(toPoint(world));
        return fromRect(iv);
    }
    getSeriesHighValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.i1(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesHighValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.v8(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    getSeriesLowValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.i3(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesLowValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.wa(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    getSeriesValuePositionFromSeriesPixel(mouse, useInterpolation, skipUnknowns) {
        let iv = this.i.wd(toPoint(mouse), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    getSeriesValueFromSeriesPixel(mouse, useInterpolation, skipUnknowns) {
        let iv = this.i.i6(toPoint(mouse), useInterpolation, skipUnknowns);
        return (iv);
    }
    /**
     * For a category plotted series, returns the current width of the items within the categories. This only returns a value if the items have some form of width (e.g. columns, bars, etc.) otherwise 0 is returned.
    
    *
    * ```ts
    * let span: number = this.series.getItemSpan();
    * ```
    */
    getItemSpan() {
        let iv = this.i.iw();
        return (iv);
    }
    /**
     * Gets the data values of each axis associated with the specified value type. For example, if you specify
     * mode as Maximum then it will return the maximum numerical value for each axis. For non-numerical axes it
     * will return NaN.
    
    * @param mode  * The type of value desired from the series axes.
    */
    getSeriesValueType(mode) {
        let iv = this.i.aq((mode == null ? null : mode));
        return (iv);
    }
    /**
     * Gets position of series value for specified layer value mode
    
    */
    getSeriesValueTypePosition(mode) {
        let iv = this.i.wf((mode == null ? null : mode));
        return fromPoint(iv);
    }
    getSeriesValueTypePositionFromValue(values) {
        let iv = this.i.wg(values);
        return fromPoint(iv);
    }
    /**
     * Hides any internal tooltips presented by the series, if any.
    
    */
    hideToolTips() {
        this.i.p7();
    }
    /**
     * Hides any internal tooltips presented by the series, if any.
    
    */
    hideToolTipsImmediate() {
        this.i.p8();
    }
    /**
     * Converts a point from coordinates within the series plot area to a world position within axis space.
    
    * @param seriesPoint  * The pixel location within the plot area of the series.
    *
    * ```ts
    * this.series.toWorldPosition({x:500,y:500});
    * ```
    */
    toWorldPosition(seriesPoint) {
        let iv = this.i.wh(toPoint(seriesPoint));
        return fromPoint(iv);
    }
    /**
     * Converts a rect from coordinates within the series plot area to a world position within axis space.
    
    * @param rect  * The pixel rect within the plot area of the series.
    */
    toWorldRect(rect) {
        let iv = this.i.wr(toRect(rect));
        return fromRect(iv);
    }
    /**
     * Converts a point from world coordinates to coordinates within the viewport of the series.
    
    * @param world  * The world position from which to convert.
    */
    fromWorldPosition(world) {
        let iv = this.i.v7(toPoint(world));
        return fromPoint(iv);
    }
    /**
     * Requests that the provided item should be brought into view if possible.
    
    * @param item  * The item to attempt to bring into view.
    *
    * ```ts
    * this.series.scrollIntoView(dataItem);
    * ```
    */
    scrollIntoView(item) {
        let iv = this.i.gd(item);
        return (iv);
    }
    /**
     * Renders the series.
    
    * @param animate  * True if the change should be animated.
    *
    * ```ts
    * this.series.renderSeries(true);
    * ```
    */
    renderSeries(animate) {
        this.i.rr(animate);
    }
    getMainContentViewport() {
        let iv = this.i.getMainContentViewport();
        return fromRect(iv);
    }
    /**
     * Gets the effective viewport, adjusted to account for margins and other factors.
    
    */
    getEffectiveViewport() {
        let iv = this.i.getEffectiveViewport();
        return fromRect(iv);
    }
    /**
     * Removes all alternate views of this series.
    
    */
    removeAllAlternateViews() {
        this.i.rk();
    }
    /**
     * Removes an alternate view of this series.
    
    * @param viewIdentifier  * The string identifier for the view to remove.
    */
    removeAlternateView(viewIdentifier) {
        this.i.rm(viewIdentifier);
    }
    /**
     * Called to notify about changes to indexed-based properties, e.g. Brushes, Outlines, MarkerBrushes, MarkerOutlines and refresh series
    
    */
    notifyIndexedPropertiesChanged() {
        this.i.qg();
    }
    /**
     * Notifies the series that a visual property has changed, requiring a visual update.
    
    */
    notifyVisualPropertiesChanged() {
        this.i.qn();
    }
    /**
     * Replays the assigned transition in animation, if any.
    
    *
    * ```ts
    * this.series.replayTransitionIn();
    * ```
    */
    replayTransitionIn() {
        this.i.rv();
    }
    /**
     * Plays the assigned transition out animation, if any.
    
    */
    playTransitionOut() {
        this.i.q7();
    }
    /**
     * Plays the assigned transition out animation, if any.
    
    */
    playTransitionIn() {
        this.i.q6();
    }
    /**
     * Plays the assigned transition out animation, if any.
    
    */
    playTransitionOutAndRemove() {
        this.i.q8();
    }
    /**
     * Call to null out the axes of the series;
    
    */
    removeAxes() {
        this.i.rn();
    }
    /**
     * Simulates a pointer hover over the series surface.
    
    * @param point  * The pointer position relative to the series viewport over which to hover.
    *
    * When tooltips are enabled, `SimulateHover` can be used to automatically display the tooltip over a particular series item.
    *
    * ```ts
    * this.series.simulateHover({x:250,y:250});
    * ```
    */
    simulateHover(point) {
        this.i.r6(toPoint(point));
    }
    /**
     * Called when this series' Style is updated.
    
    */
    styleUpdated() {
        this.i.sc();
    }
    setNegativeColors(negativeBrush, negativeOutline) {
        this.i.r3(stringToBrush(negativeBrush), stringToBrush(negativeOutline));
    }
    /**
     * Moves the cursor point to the provided world position. Some series may react to the current cursor position.
    
    * @param point  * The cursor point, in world coordinates.
    *
    * `MoveCursorPoint` can be used on annotation layers to programitically show the annotation at the given point.
    *
    * ```ts
    * this.crosshair.moveCursorPoint({x:.5,y:.5});
    * ```
    */
    moveCursorPoint(point) {
        this.i.qc(toPoint(point));
    }
    /**
     * Fired when a transition out has completed.
    */
    get transitionOutCompleted() {
        return this._transitionOutCompleted;
    }
    set transitionOutCompleted(ev) {
        if (this._transitionOutCompleted_wrapped !== null) {
            this.i.transitionOutCompleted = delegateRemove(this.i.transitionOutCompleted, this._transitionOutCompleted_wrapped);
            this._transitionOutCompleted_wrapped = null;
            this._transitionOutCompleted = null;
        }
        this._transitionOutCompleted = ev;
        this._transitionOutCompleted_wrapped = (o, e) => {
            let outerArgs = new IgrTransitionOutCompletedEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeTransitionOutCompleted) {
                this.beforeTransitionOutCompleted(this, outerArgs);
            }
            if (this._transitionOutCompleted) {
                this._transitionOutCompleted(this, outerArgs);
            }
        };
        this.i.transitionOutCompleted = delegateCombine(this.i.transitionOutCompleted, this._transitionOutCompleted_wrapped);
        ;
    }
    /**
     * This event is raised every time a render has been requested from a series.
    *
    * ```ts
    *  const series1 = new IgrAreaSeries({ name: "series1" });
    * series1.valueMemberPath = "USA";
    * series1.title = "USA";
    * series1.xAxisName = "xAxis";
    * series1.yAxisName = "yAxis";
    * series1.renderRequested= this.onRenderRequested;
    *
    * public onRenderRequested =( s:IgrSeries, e: RenderRequestedEventArgs) => {
    *           }
    * ```
    *
    * ```ts
    *  <IgrDataChart
    *     dataSource={this.state.dataSource} >
    *
    *     <IgrNumericXAxis name="xAxis" />
    *     <IgrNumericYAxis name="yAxis" />
    *
    *     <IgrAreaSeries
    *      name="series1"
    *     xAxisName="xAxis"
    * 	yAxisName="yAxis"
    * 	valueMemberPath="Value"
    *     shadowBlur="7"
    *     shadowColor="blue"
    * 	renderRequested= {this.onRenderRequested}/>
    *  </IgrDataChart>
    * ```
    */
    get renderRequested() {
        return this._renderRequested;
    }
    set renderRequested(ev) {
        if (this._renderRequested_wrapped !== null) {
            this.i.renderRequested = delegateRemove(this.i.renderRequested, this._renderRequested_wrapped);
            this._renderRequested_wrapped = null;
            this._renderRequested = null;
        }
        this._renderRequested = ev;
        this._renderRequested_wrapped = (o, e) => {
            let outerArgs = new IgrRenderRequestedEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeRenderRequested) {
                this.beforeRenderRequested(this, outerArgs);
            }
            if (this._renderRequested) {
                this._renderRequested(this, outerArgs);
            }
        };
        this.i.renderRequested = delegateCombine(this.i.renderRequested, this._renderRequested_wrapped);
        ;
    }
}
