// igniteui-react-charts/esm2015/lib/igr-data-legend-summary-event-args.js
/**
 * Represents event argument for DataLegendSummaryCalculationCustom
*/
export class IgrDataLegendSummaryEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * Gets actual values in the current a column
    */
    get columnValues() {
        return this.i.columnValues;
    }
    set columnValues(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.columnValues = v;
    }
    /**
     * Gets member path to identify a column
    */
    get columnMemberPath() {
        return this.i.columnMemberPath;
    }
    set columnMemberPath(v) {
        this.i.columnMemberPath = v;
    }
    /**
     * Gets or sets summary values displayed below data column
    */
    get summaryValue() {
        return this.i.summaryValue;
    }
    set summaryValue(v) {
        this.i.summaryValue = +v;
    }
    /**
     * Gets or sets text displayed or right side of SummaryValue
    */
    get summaryUnits() {
        return this.i.summaryUnits;
    }
    set summaryUnits(v) {
        this.i.summaryUnits = v;
    }
    /**
     * Gets or sets text displayed or left side of SummaryValue
    */
    get summaryLabel() {
        return this.i.summaryLabel;
    }
    set summaryLabel(v) {
        this.i.summaryLabel = v;
    }
}
