// igniteui-react-charts/esm2015/lib/igr-data-source-supporting-calculation.js
/**
 * Represents a calculation strategy that uses the calculation data source
 * to product a column of values.
*
* `DataSourceSupportingCalculation` class Represents a calculation strategy that uses the calculation data source to product a column of values.
*
*     DataSourceSupportingCalculation SC = new DataSourceSupportingCalculation();
*/
export class IgrDataSourceSupportingCalculation {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
