// igniteui-react-charts/esm2015/lib/igr-label-click-event-args.js
import { ensureBool } from "igniteui-react-core";
/**
 * Contains PieChart label click event data.
*/
export class IgrLabelClickEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * Gets the slice data context.
    */
    get item() {
        return this.i.item;
    }
    /**
     * Gets and sets whether or not the owning pie chart should fire its SliceClick event
    */
    get allowSliceClick() {
        return this.i.allowSliceClick;
    }
    set allowSliceClick(v) {
        this.i.allowSliceClick = ensureBool(v);
    }
}
