// igniteui-react-charts/esm2015/lib/igr-anchored-radial-series.js
import { TrendLineType_$type } from "igniteui-react-core";
import { IgrRadialBase } from "./igr-radial-base";
import { ensureBool, ensureEnum, brushToString, stringToBrush, toDoubleCollection, fromDoubleCollection, toPoint, fromPoint } from "igniteui-react-core";
/**
 * Represents the base class for all IgxDataChartComponent anchored radial category series.
*/
export class IgrAnchoredRadialSeries extends IgrRadialBase {
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or sets whether to normalize the values against the category values if using a proportional category axis.
    */
    get useCategoryNormalizedValues() {
        return this.i.ach;
    }
    set useCategoryNormalizedValues(v) {
        this.i.ach = ensureBool(v);
    }
    /**
     * Gets or sets the item path that provides the values for the current series.
    */
    get valueMemberPath() {
        return this.i.acx;
    }
    set valueMemberPath(v) {
        this.i.acx = v;
    }
    /**
     * Gets or sets the value mapping property for the current series object.
    */
    get highlightedValueMemberPath() {
        return this.i.acp;
    }
    set highlightedValueMemberPath(v) {
        this.i.acp = v;
    }
    /**
     * Gets or sets the label displayed before series value in the Data Legend.
    */
    get valueMemberAsLegendLabel() {
        return this.i.act;
    }
    set valueMemberAsLegendLabel(v) {
        this.i.act = v;
    }
    /**
     * Gets or sets the unit displayed after series value in the Data Legend.
    */
    get valueMemberAsLegendUnit() {
        return this.i.acv;
    }
    set valueMemberAsLegendUnit(v) {
        this.i.acv = v;
    }
    /**
     * Gets or sets the trend type for the current series object.
    */
    get trendLineType() {
        return this.i.ab6;
    }
    set trendLineType(v) {
        this.i.ab6 = ensureEnum(TrendLineType_$type, v);
    }
    /**
     * Gets or sets the brush that specifies how to the current series
     * object's Trend line is drawn.
    */
    get trendLineBrush() {
        return brushToString(this.i.adl);
    }
    set trendLineBrush(v) {
        this.i.adl = stringToBrush(v);
    }
    /**
     * Gets the effective TrendLineBrush for this series.
    */
    get actualTrendLineBrush() {
        return brushToString(this.i.adk);
    }
    set actualTrendLineBrush(v) {
        this.i.adk = stringToBrush(v);
    }
    /**
     * Gets or sets the thickness of the current series object's trend line.
    */
    get trendLineThickness() {
        return this.i.acm;
    }
    set trendLineThickness(v) {
        this.i.acm = +v;
    }
    /**
     * Gets or sets a collection of double values that indicate the pattern of dashes and gaps that
     * is used to draw the trend line for the current series object.
    */
    get trendLineDashArray() {
        return fromDoubleCollection(this.i.adm);
    }
    set trendLineDashArray(v) {
        this.i.adm = toDoubleCollection(v);
    }
    /**
     * Gets or sets the trend line period for the current series.
     * The typical, and initial, value for bollinger band periods is 20.
    */
    get trendLinePeriod() {
        return this.i.acn;
    }
    set trendLinePeriod(v) {
        this.i.acn = +v;
    }
    /**
     * Sets or Gets the Z index of the trendline.
    */
    get trendLineZIndex() {
        return this.i.aco;
    }
    set trendLineZIndex(v) {
        this.i.aco = +v;
    }
    get hasValueAxis() {
        return this.i.ek;
    }
    get isValueAxisInverted() {
        return this.i.f5;
    }
    getItemValue(item, memberPathName) {
        let iv = this.i.kq(item, memberPathName);
        return (iv);
    }
    /**
     * Gets the value of a requested member path from the series.
    
    * @param memberPathName  * The property name of a valid member path for the series
    */
    getMemberPathValue(memberPathName) {
        let iv = this.i.md(memberPathName);
        return (iv);
    }
    getPreviousOrExactIndex(world, skipUnknowns) {
        let iv = this.i.ka(toPoint(world), skipUnknowns);
        return (iv);
    }
    getNextOrExactIndex(world, skipUnknowns) {
        let iv = this.i.j8(toPoint(world), skipUnknowns);
        return (iv);
    }
    getSeriesValue(world, useInterpolation, skipUnknowns) {
        let iv = this.i.i5(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    }
    getSeriesValuePosition(world, useInterpolation, skipUnknowns) {
        let iv = this.i.wc(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    }
    /**
     * Returns the offset value for this series if grouped on a category axis.
    
    */
    getOffsetValue() {
        let iv = this.i.getOffsetValue();
        return (iv);
    }
    /**
     * Returns the width of the category grouping this series is in.
    
    */
    getCategoryWidth() {
        let iv = this.i.getCategoryWidth();
        return (iv);
    }
    /**
     * Scrolls the specified item into the view.
    
    * @param item  * The item to scroll into view.
    */
    scrollIntoView(item) {
        let iv = this.i.gd(item);
        return (iv);
    }
}
