// igniteui-react-charts/esm2015/lib/igr-value-brush-scale.js
import { IgrBrushScale } from "./igr-brush-scale";
import { ValueBrushScale } from "./ValueBrushScale";
import { ensureBool } from "igniteui-react-core";
/**
 * Represents a brush scale that uses value-based brush selection.
*/
export class IgrValueBrushScale extends IgrBrushScale {
    createImplementation() {
        return new ValueBrushScale();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets or sets the minimum value for this scale.
    */
    get minimumValue() {
        return this.i.minimumValue;
    }
    set minimumValue(v) {
        this.i.minimumValue = +v;
    }
    /**
     * Gets or sets the maximum value for this scale.
    */
    get maximumValue() {
        return this.i.maximumValue;
    }
    set maximumValue(v) {
        this.i.maximumValue = +v;
    }
    /**
     * Gets or sets whether the scale is logarithmic.
    */
    get isLogarithmic() {
        return this.i.isLogarithmic;
    }
    set isLogarithmic(v) {
        this.i.isLogarithmic = ensureBool(v);
    }
    /**
     * Gets or sets the logarithm base for this scale.
    */
    get logarithmBase() {
        return this.i.logarithmBase;
    }
    set logarithmBase(v) {
        this.i.logarithmBase = +v;
    }
    /**
     * Checks if the value brush scale is ready for usage in the chart
    */
    get isReady() {
        return this.i.isReady;
    }
}
