// igniteui-react-charts/esm2015/lib/igr-financial-event-args.js
import { IgrFinancialCalculationDataSource } from './igr-financial-calculation-data-source';
import { IgrFinancialCalculationSupportingCalculations } from './igr-financial-calculation-supporting-calculations';
import { List$1 } from "igniteui-react-core";
import { String_$type } from "igniteui-react-core";
/**
 * Parameterizes a financial calculation event.
*/
export class IgrFinancialEventArgs {
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    get basedOn() {
        let ret = null;
        if (this.i.basedOn) {
            ret = [];
            for (let i = 0; i < this.i.basedOn.count; i++) {
                ret.push(this.i.basedOn[i]);
            }
        }
        return ret;
    }
    set basedOn(v) {
        if (!v) {
            this.i.basedOn = null;
            return;
        }
        let list = new List$1(String_$type, 0);
        for (let i = 0; i < v.length; i++) {
            list.add(v[i]);
        }
        this.i.basedOn = list;
    }
    /**
 * The beginning position that should be calculated from.
*/
    get position() {
        return this.i.position;
    }
    set position(v) {
        this.i.position = +v;
    }
    /**
     * The number of positions that should be calculated from the start.
    */
    get count() {
        return this.i.count;
    }
    set count(v) {
        this.i.count = +v;
    }
    /**
     * The data to use for the calculation.
    */
    get dataSource() {
        const r = this.i.dataSource;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrFinancialCalculationDataSource();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set dataSource(v) {
        v == null ? this.i.dataSource = null : this.i.dataSource = v.i;
    }
    /**
     * The supporting calculations to use in the calculation.
    */
    get supportingCalculations() {
        const r = this.i.supportingCalculations;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrFinancialCalculationSupportingCalculations();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set supportingCalculations(v) {
        v == null ? this.i.supportingCalculations = null : this.i.supportingCalculations = v.i;
    }
}
