// igniteui-react-charts/esm2015/lib/igr-plot-area-mouse-event-args.js
import { fromPoint, toPoint, ensureBool } from "igniteui-react-core";
/**
 * Provides data for IgxDataChartComponent mouse button related events.
*/
export class IgrPlotAreaMouseEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get chartPosition() {
        return fromPoint(this.i.chartPosition);
    }
    set chartPosition(v) {
        this.i.chartPosition = toPoint(v);
    }
    get plotAreaPosition() {
        return fromPoint(this.i.plotAreaPosition);
    }
    set plotAreaPosition(v) {
        this.i.plotAreaPosition = toPoint(v);
    }
    get isDuringManipulation() {
        return this.i.isDuringManipulation;
    }
    set isDuringManipulation(v) {
        this.i.isDuringManipulation = ensureBool(v);
    }
    /**
     * Gets the Chart associated with the current event.
    */
    get viewer() {
        const r = this.i.viewer;
        if (r == null) {
            return null;
        }
        return r.externalObject;
    }
    set viewer(v) {
        v == null ? this.i.viewer = null : this.i.viewer = v.i;
    }
}
