// igniteui-react-charts/esm2015/lib/ItemwiseStrategyBasedIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ItemwiseStrategyCalculationStrategy } from "./ItemwiseStrategyCalculationStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ItemwiseStrategyBasedIndicator = /*@__PURE__*/ (() => {
    class ItemwiseStrategyBasedIndicator extends StrategyBasedIndicator {
        get_abv() {
            return new ItemwiseStrategyCalculationStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        aci() {
            return true;
        }
        constructor() {
            super();
        }
        ab9() {
            super.ab9();
            this.acg = this.ach;
            this.abu.c = this.acg;
        }
    }
    ItemwiseStrategyBasedIndicator.$t = /*@__PURE__*/ markType(ItemwiseStrategyBasedIndicator, 'ItemwiseStrategyBasedIndicator', StrategyBasedIndicator.$);
    return ItemwiseStrategyBasedIndicator;
})();
