// igniteui-react-charts/esm2015/lib/ItemsSourceAction.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available actions that can cause an ItemsSource event.
 */
export var ItemsSourceAction = /*@__PURE__*/ (function (ItemsSourceAction) {
    /**
     * One or more rows were removed from the data source.
     */
    ItemsSourceAction[ItemsSourceAction["Remove"] = 0] = "Remove";
    /**
     * One or more rows were added to the data source.
     */
    ItemsSourceAction[ItemsSourceAction["Insert"] = 1] = "Insert";
    /**
     * One or more rows were replaced in the data source.
     */
    ItemsSourceAction[ItemsSourceAction["Replace"] = 2] = "Replace";
    /**
     * An item property value changed in the data source.
     */
    ItemsSourceAction[ItemsSourceAction["Change"] = 3] = "Change";
    /**
     * The entire data source contents were reset.
     */
    ItemsSourceAction[ItemsSourceAction["Reset"] = 4] = "Reset";
    return ItemsSourceAction;
})({});
/**
 * @hidden
 */
export let ItemsSourceAction_$type = /*@__PURE__*/ markEnum('ItemsSourceAction', 'Remove,0|Insert,1|Replace,2|Change,3|Reset,4');
