// igniteui-react-charts/esm2015/lib/ItemTooltipCollisionInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ItemTooltipCollisionInfo = /*@__PURE__*/ (() => {
    class ItemTooltipCollisionInfo extends Base {
        constructor() {
            super(...arguments);
            this.e = null;
            this.d = 0;
            this.c = 0;
            this.a = 0;
            this.b = 0;
        }
    }
    ItemTooltipCollisionInfo.$t = /*@__PURE__*/ markType(ItemTooltipCollisionInfo, 'ItemTooltipCollisionInfo');
    return ItemTooltipCollisionInfo;
})();
