// igniteui-react-charts/esm2015/lib/KDTreeNode2D.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let KDTreeNode2D = /*@__PURE__*/ (() => {
    class KDTreeNode2D extends Base {
        constructor() {
            super(...arguments);
            this.g = false;
            this.f = false;
            this.h = 0;
            this.b = null;
            this.c = null;
            this.d = null;
            this.a = null;
            this.e = null;
        }
    }
    KDTreeNode2D.$t = /*@__PURE__*/ markType(KDTreeNode2D, 'KDTreeNode2D');
    return KDTreeNode2D;
})();
