// igniteui-react-charts/esm2015/lib/KNearestResults.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let KNearestResults = /*@__PURE__*/ (() => {
    class KNearestResults extends Base {
        constructor() {
            super(...arguments);
            this.e = 0;
            this.f = 0;
            this.b = null;
            this.a = false;
            this.d = 0;
            this.c = 0;
            this.g = 0;
        }
    }
    KNearestResults.$t = /*@__PURE__*/ markType(KNearestResults, 'KNearestResults');
    return KNearestResults;
})();
