// igniteui-react-charts/esm2015/lib/LabelsPosition.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available positions of labels.
 */
export var LabelsPosition = /*@__PURE__*/ (function (LabelsPosition) {
    /**
     * No labels will be displayed.
     */
    LabelsPosition[LabelsPosition["None"] = 0] = "None";
    /**
     * Labels will be displayed in the center.
     */
    LabelsPosition[LabelsPosition["Center"] = 1] = "Center";
    /**
     * Labels will be displayed inside and by the edge of the container.
     */
    LabelsPosition[LabelsPosition["InsideEnd"] = 2] = "InsideEnd";
    /**
     * Labels will be displayed outside the container.
     */
    LabelsPosition[LabelsPosition["OutsideEnd"] = 3] = "OutsideEnd";
    /**
     * Labels will automatically decide their location.
     */
    LabelsPosition[LabelsPosition["BestFit"] = 4] = "BestFit";
    return LabelsPosition;
})({});
/**
 * @hidden
 */
export let LabelsPosition_$type = /*@__PURE__*/ markEnum('LabelsPosition', 'None,0|Center,1|InsideEnd,2|OutsideEnd,3|BestFit,4');
