// igniteui-react-charts/esm2015/lib/LabelFontHeuristics.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let LabelFontHeuristics = /*@__PURE__*/ (() => {
    class LabelFontHeuristics extends Base {
        constructor() {
            super();
            this.b = 0;
            this.e = 0;
            this.c = 0;
            this.d = 0;
            this.a = false;
            this.b = NaN;
            this.e = NaN;
            this.c = NaN;
            this.d = NaN;
        }
    }
    LabelFontHeuristics.$t = /*@__PURE__*/ markType(LabelFontHeuristics, 'LabelFontHeuristics');
    return LabelFontHeuristics;
})();
