// igniteui-react-charts/esm2015/lib/LegendVisualData.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { LegendVisualDataList } from "./LegendVisualDataList";
import { StringBuilder } from "igniteui-react-core";
/**
 * @hidden
 */
export let LegendVisualData = /*@__PURE__*/ (() => {
    class LegendVisualData extends Base {
        constructor() {
            super();
            this._items = null;
            this._viewport = null;
            this._width = 0;
            this._height = 0;
            this.items = new LegendVisualDataList();
        }
        get items() {
            return this._items;
        }
        set items(a) {
            this._items = a;
        }
        get viewport() {
            return this._viewport;
        }
        set viewport(a) {
            this._viewport = a;
        }
        get width() {
            return this._width;
        }
        set width(a) {
            this._width = a;
        }
        get height() {
            return this._height;
        }
        set height(a) {
            this._height = a;
        }
        serialize() {
            let a = new StringBuilder(0);
            a.u("{");
            a.u("width: " + this.width + ",");
            a.u("height: " + this.height + ",");
            a.u("items: [");
            for (let b = 0; b < this.items.count; b++) {
                if (b != 0) {
                    a.l(", ");
                }
                a.l(this.items._inner[b].serialize());
            }
            a.u("],");
            a.u("}");
            return a.toString();
        }
    }
    LegendVisualData.$t = /*@__PURE__*/ markType(LegendVisualData, 'LegendVisualData');
    return LegendVisualData;
})();
