// igniteui-react-charts/esm2015/lib/LegendMouseEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { EventArgs, typeCast, markType } from "igniteui-react-core";
import { ILegendSeriesInternal_$type } from "igniteui-react-core";
/**
 * @hidden
 */
export let LegendMouseEventArgs = /*@__PURE__*/ (() => {
    class LegendMouseEventArgs extends EventArgs {
        constructor(a, b, c, d, e) {
            super();
            this._originalEvent = null;
            this._item = null;
            this._series = null;
            this._chart = null;
            this._legendItem = null;
            this.chart = a;
            this.series = b;
            this.item = c;
            this.originalEvent = d;
            this.legendItem = e;
        }
        toString() {
            return this.chart.name + ", " + this.series.name + ", " + (this.item != null ? this.item.toString() : "") + ", " + this.getPosition(null).toString();
        }
        get originalEvent() {
            return this._originalEvent;
        }
        set originalEvent(a) {
            this._originalEvent = a;
        }
        getPosition(a) {
            return this.originalEvent.getPosition(a);
        }
        get originalSource() {
            return this.originalEvent.originalSource;
        }
        get item() {
            return this._item;
        }
        set item(a) {
            this._item = a;
        }
        get series() {
            return this._series;
        }
        set series(a) {
            this._series = a;
        }
        get d() {
            return typeCast(ILegendSeriesInternal_$type, this.series);
        }
        get chart() {
            return this._chart;
        }
        set chart(a) {
            this._chart = a;
        }
        get legendItem() {
            return this._legendItem;
        }
        set legendItem(a) {
            this._legendItem = a;
        }
    }
    LegendMouseEventArgs.$t = /*@__PURE__*/ markType(LegendMouseEventArgs, 'LegendMouseEventArgs', EventArgs.$);
    return LegendMouseEventArgs;
})();
