// igniteui-react-charts/esm2015/lib/LogarithmicScaler.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NumericScaler } from "./NumericScaler";
import { markType } from "igniteui-react-core";
import { AutoRangeCalculator } from "./AutoRangeCalculator";
/**
 * @hidden
 */
export let LogarithmicScaler = /*@__PURE__*/ (() => {
    class LogarithmicScaler extends NumericScaler {
        constructor() {
            super(...arguments);
            this.ae = 0;
            this.ad = 0;
        }
        w(a, b, c) {
            super.w(a, b, c);
            switch (a) {
                case NumericScaler.$$p[1]:
                    this.ae = Math.log(this.l);
                    break;
                case NumericScaler.$$p[0]:
                    this.ad = Math.log(this.k);
                    break;
            }
        }
        t(a, b, c, d, e) {
            let f;
            let g;
            let h = AutoRangeCalculator.a(a, b, c, true, a.mm, f, g);
            f = h.p5;
            g = h.p6;
            d = f;
            e = g;
            return {
                p3: d,
                p4: e
            };
        }
    }
    LogarithmicScaler.$t = /*@__PURE__*/ markType(LogarithmicScaler, 'LogarithmicScaler', NumericScaler.$);
    return LogarithmicScaler;
})();
