// igniteui-react-charts/esm2015/lib/Marker.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ContentControl } from "igniteui-react-core";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let Marker = /*@__PURE__*/ (() => {
    class Marker extends ContentControl {
        constructor() {
            super(...arguments);
            this.am = null;
            this.an = null;
            this.ak = 0;
            this.al = 0;
            this.ai = 0;
            this.aj = 0;
        }
    }
    Marker.$t = /*@__PURE__*/ markType(Marker, 'Marker', ContentControl.$);
    return Marker;
})();
