// igniteui-react-charts/esm2015/lib/MarkerTemplates.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { CanvasViewRenderer } from "igniteui-react-core";
/**
 * @hidden
 */
export let MarkerTemplates = /*@__PURE__*/ (() => {
    class MarkerTemplates extends Base {
        static n(a, b, c) {
            a.beginPath();
            let d = b.actualItemBrush;
            let e = b.outline;
            if (d != null) {
                if (d.isGradient) {
                    let f = c.availableWidth / 2;
                    let g = c.availableHeight / 2;
                    let h = c.xPosition - f;
                    let i = c.yPosition - f;
                    h += c.renderOffsetX;
                    i += c.renderOffsetY;
                    let j = CanvasViewRenderer.k(a, d, h, i, h + c.availableWidth, i + c.availableHeight, b.thickness, false);
                    a.fillStyle = j;
                }
                else {
                    if (a.fillStyle != d._fill) {
                        a.fillStyle = d._fill;
                    }
                }
            }
            if (b.outline != null) {
                if (b.outline.isGradient) {
                    let k = c.availableWidth / 2;
                    let l = c.availableHeight / 2;
                    let m = c.xPosition - k;
                    let n = c.yPosition - k;
                    m += c.renderOffsetX;
                    n += c.renderOffsetY;
                    let o = CanvasViewRenderer.k(a, b.outline, m, n, m + c.availableWidth, n + c.availableHeight, b.thickness, true);
                    a.strokeStyle = o;
                }
                else {
                    if (a.strokeStyle != b.outline._fill) {
                        a.strokeStyle = b.outline._fill;
                    }
                }
                a.lineWidth = b.thickness;
            }
        }
        static a(a, b) {
            a.closePath();
            if (b.actualItemBrush != null) {
                a.fill();
            }
            if (b.outline != null) {
                a.stroke();
            }
        }
        static b(a) {
            a.isConstant = true;
            a.width = 8;
            a.height = 8;
        }
        static d(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = Math.min(a.availableWidth, a.availableHeight) / 2;
            b.arc(a.xPosition, a.yPosition, d, 0, (Math.PI * 2), false);
            MarkerTemplates.a(b, c);
        }
        static m(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = (a.availableHeight / 2);
            let e = (a.availableWidth / 2);
            let f = (a.xPosition - e);
            let g = (a.yPosition - d);
            b.moveTo(f, g);
            b.lineTo(f + e, g + a.availableHeight);
            b.lineTo(f + a.availableWidth, g + 0);
            b.lineTo(f, g);
            MarkerTemplates.a(b, c);
        }
        static j(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = (a.availableHeight / 2);
            let e = (a.availableWidth / 2);
            let f = (a.xPosition - e);
            let g = (a.yPosition - d);
            b.moveTo(f, g + a.availableHeight);
            b.lineTo(f + e, g + 0);
            b.lineTo(f + a.availableWidth, g + a.availableHeight);
            b.lineTo(f, g + a.availableHeight);
            MarkerTemplates.a(b, c);
        }
        static k(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = (a.availableHeight / 2);
            let e = (a.availableWidth / 2);
            let f = (a.xPosition - e);
            let g = (a.yPosition - d);
            b.rect(f, g, a.availableWidth, a.availableHeight);
            MarkerTemplates.a(b, c);
        }
        static c(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = (a.availableHeight / 2);
            let e = (a.availableWidth / 2);
            let left_ = (a.xPosition - e);
            let top_ = (a.yPosition - d);
            b.rect((left_ | 0), (top_ | 0), a.availableWidth, a.availableHeight);
            MarkerTemplates.a(b, c);
        }
        static e(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = (a.availableHeight / 2);
            let e = (a.availableWidth / 2);
            let f = (a.xPosition - e);
            let g = (a.yPosition - d);
            b.moveTo(f + 0, g + d);
            b.lineTo(f + e, g + a.availableHeight);
            b.lineTo(f + a.availableWidth, g + d);
            b.lineTo(f + e, g + 0);
            b.lineTo(f + 0, g + d);
            MarkerTemplates.a(b, c);
        }
        static h(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = (a.availableHeight / 2);
            let e = (a.availableWidth / 2);
            let f = (a.xPosition - e);
            let g = (a.yPosition - d);
            b.moveTo(f + ((4 / 8) * a.availableWidth), g + 0);
            b.lineTo(f + ((0.1956000030040741 / 8) * a.availableWidth), g + ((2.7639999389648438 / 8) * a.availableHeight));
            b.lineTo(f + ((1.6499999761581421 / 8) * a.availableWidth), g + ((7.2360000610351563 / 8) * a.availableHeight));
            b.lineTo(f + ((6.3499999046325684 / 8) * a.availableWidth), g + ((7.2360000610351563 / 8) * a.availableHeight));
            b.lineTo(f + ((7.8043999671936035 / 8) * a.availableWidth), g + ((2.7639999389648438 / 8) * a.availableHeight));
            b.lineTo(f + ((4 / 8) * a.availableWidth), g + 0);
            MarkerTemplates.a(b, c);
        }
        static f(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = (a.availableHeight / 2);
            let e = (a.availableWidth / 2);
            let f = (a.xPosition - e);
            let g = (a.yPosition - d);
            b.moveTo(f + ((4 / 8) * a.availableWidth), g + 0);
            b.lineTo(f + ((0.35600000619888306 / 8) * a.availableWidth), g + ((2 / 8) * a.availableHeight));
            b.lineTo(f + ((0.35600000619888306 / 8) * a.availableWidth), g + ((6 / 8) * a.availableHeight));
            b.lineTo(f + ((4 / 8) * a.availableWidth), g + (a.availableHeight));
            b.lineTo(f + ((7.4640002250671387 / 8) * a.availableWidth), g + ((6 / 8) * a.availableHeight));
            b.lineTo(f + ((7.4640002250671387 / 8) * a.availableWidth), g + ((2 / 8) * a.availableHeight));
            b.lineTo(f + ((4 / 8) * a.availableWidth), g + 0);
            MarkerTemplates.a(b, c);
        }
        static l(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = (a.availableHeight / 2);
            let e = (a.availableWidth / 2);
            let f = (a.xPosition - e);
            let g = (a.yPosition - d);
            b.moveTo(f + ((4 / 8) * a.availableWidth), g + 0);
            b.lineTo(f + ((2.5855998992919922 / 8) * a.availableWidth), g + ((2.5856 / 8) * a.availableHeight));
            b.lineTo(f + ((0 / 8) * a.availableWidth), g + ((4 / 8) * a.availableHeight));
            b.lineTo(f + ((2.5855998992919922 / 8) * a.availableWidth), g + ((5.4144001007080078 / 8) * a.availableHeight));
            b.lineTo(f + ((4 / 8) * a.availableWidth), g + ((8 / 8) * a.availableHeight));
            b.lineTo(f + ((5.4144001007080078 / 8) * a.availableWidth), g + ((5.4144001007080078 / 8) * a.availableHeight));
            b.lineTo(f + ((8 / 8) * a.availableWidth), g + ((4 / 8) * a.availableHeight));
            b.lineTo(f + ((5.4144001007080078 / 8) * a.availableWidth), g + ((2.5855998992919922 / 8) * a.availableHeight));
            b.lineTo(f + ((4 / 8) * a.availableWidth), g + 0);
            MarkerTemplates.a(b, c);
        }
        static i(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = (a.availableHeight / 2);
            let e = (a.availableWidth / 2);
            let f = (a.xPosition - e);
            let g = (a.yPosition - d);
            b.moveTo(f + ((4 / 8) * a.availableWidth), g + 0);
            b.lineTo(f + ((2.8243999481201172 / 8) * a.availableWidth), g + ((2.382 / 8) * a.availableHeight));
            b.lineTo(f + ((0.1956 / 8) * a.availableWidth), g + ((2.7639999389648438 / 8) * a.availableHeight));
            b.lineTo(f + ((2.0980000495910645 / 8) * a.availableWidth), g + ((4.6180000305175781 / 8) * a.availableHeight));
            b.lineTo(f + ((1.6488000154495239 / 8) * a.availableWidth), g + ((7.2360000610351563 / 8) * a.availableHeight));
            b.lineTo(f + ((4 / 8) * a.availableWidth), g + ((6 / 8) * a.availableHeight));
            b.lineTo(f + ((6.3512 / 8) * a.availableWidth), g + ((7.2360000610351563 / 8) * a.availableHeight));
            b.lineTo(f + ((5.9019999504089355 / 8) * a.availableWidth), g + ((4.6180000305175781 / 8) * a.availableHeight));
            b.lineTo(f + ((7.8043999671936035 / 8) * a.availableWidth), g + ((2.7639999389648438 / 8) * a.availableHeight));
            b.lineTo(f + ((5.1756000518798828 / 8) * a.availableWidth), g + ((2.3819999694824219 / 8) * a.availableHeight));
            b.lineTo(f + ((4 / 8) * a.availableWidth), g + 0);
            MarkerTemplates.a(b, c);
        }
        static g(a) {
            let b = a.context;
            let c = a.data;
            MarkerTemplates.n(b, c, a);
            let d = (a.availableHeight / 2);
            let e = (a.availableWidth / 2);
            let f = (a.xPosition - e);
            let g = (a.yPosition - d);
            b.moveTo(f + ((4 / 8) * a.availableWidth), g + 0);
            b.lineTo(f + ((3 / 8) * a.availableWidth), g + ((2.268 / 8) * a.availableHeight));
            b.lineTo(f + ((0.536 / 8) * a.availableWidth), g + ((2 / 8) * a.availableHeight));
            b.lineTo(f + ((2 / 8) * a.availableWidth), g + ((4 / 8) * a.availableHeight));
            b.lineTo(f + ((0.536 / 8) * a.availableWidth), g + ((6 / 8) * a.availableHeight));
            b.lineTo(f + ((3 / 8) * a.availableWidth), g + ((5.73199987411499 / 8) * a.availableHeight));
            b.lineTo(f + ((4 / 8) * a.availableWidth), g + ((8 / 8) * a.availableHeight));
            b.lineTo(f + ((5 / 8) * a.availableWidth), g + ((5.73199987411499 / 8) * a.availableHeight));
            b.lineTo(f + ((7.4640002250671387 / 8) * a.availableWidth), g + ((6 / 8) * a.availableHeight));
            b.lineTo(f + ((6 / 8) * a.availableWidth), g + ((4 / 8) * a.availableHeight));
            b.lineTo(f + ((7.4640002250671387 / 8) * a.availableWidth), g + ((2 / 8) * a.availableHeight));
            b.lineTo(f + ((5 / 8) * a.availableWidth), g + ((2.2679998874664307 / 8) * a.availableHeight));
            b.lineTo(f + ((4 / 8) * a.availableWidth), g + 0);
            MarkerTemplates.a(b, c);
        }
    }
    MarkerTemplates.$t = /*@__PURE__*/ markType(MarkerTemplates, 'MarkerTemplates');
    return MarkerTemplates;
})();
