// igniteui-react-charts/esm2015/lib/MarketFacilitationIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ItemwiseStrategyBasedIndicator } from "./ItemwiseStrategyBasedIndicator";
import { MarketFacilitationIndexIndicatorStrategy } from "./MarketFacilitationIndexIndicatorStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let MarketFacilitationIndexIndicator = /*@__PURE__*/ (() => {
    class MarketFacilitationIndexIndicator extends ItemwiseStrategyBasedIndicator {
        bj() {
            return new MarketFacilitationIndexIndicator();
        }
        get_aa2() {
            return 1;
        }
        constructor() {
            super();
        }
        get_ach() {
            return new MarketFacilitationIndexIndicatorStrategy();
        }
        get ach() {
            return this.get_ach();
        }
        get_ab8() {
            return MarketFacilitationIndexIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
    }
    MarketFacilitationIndexIndicator.$t = /*@__PURE__*/ markType(MarketFacilitationIndexIndicator, 'MarketFacilitationIndexIndicator', ItemwiseStrategyBasedIndicator.$);
    return MarketFacilitationIndexIndicator;
})();
