// igniteui-react-charts/esm2015/lib/MarkerMessage_Spark.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RenderingMessage_Spark } from "./RenderingMessage_Spark";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let MarkerMessage_Spark = /*@__PURE__*/ (() => {
    class MarkerMessage_Spark extends RenderingMessage_Spark {
        constructor() {
            super(...arguments);
            this.e = null;
            this.f = null;
            this.m = null;
            this.n = null;
            this.d = null;
            this.c = null;
            this.k = 0;
            this.g = 0;
            this.i = 0;
            this.h = 0;
            this.j = 0;
            this.l = 0;
        }
    }
    MarkerMessage_Spark.$t = /*@__PURE__*/ markType(MarkerMessage_Spark, 'MarkerMessage_Spark', RenderingMessage_Spark.$);
    return MarkerMessage_Spark;
})();
