// igniteui-react-charts/esm2015/lib/MarkerManagerBucket.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Number_$type, markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
/**
 * @hidden
 */
export let MarkerManagerBucket = /*@__PURE__*/ (() => {
    class MarkerManagerBucket extends Base {
        constructor() {
            super(...arguments);
            this.b = null;
            this.c = null;
        }
        get d() {
            if (this.b == null) {
                this.b = new List$1(Number_$type, 0);
            }
            return this.b;
        }
        get e() {
            if (this.c == null) {
                this.c = new List$1(Number_$type, 0);
            }
            return this.c;
        }
        f(a) {
            if (this.e.count > 0) {
                let b = this.e._inner[this.e.count - 1];
                this.e.removeAt(this.e.count - 1);
                a = true;
                return {
                    ret: b,
                    p0: a
                };
            }
            let c = this.d._inner[this.d.count - 1];
            this.d.removeAt(this.d.count - 1);
            a = false;
            return {
                ret: c,
                p0: a
            };
        }
        get a() {
            return this.d.count == 0 && this.e.count == 0;
        }
    }
    MarkerManagerBucket.$t = /*@__PURE__*/ markType(MarkerManagerBucket, 'MarkerManagerBucket');
    return MarkerManagerBucket;
})();
