// igniteui-react-charts/esm2015/lib/MassIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { MassIndexIndicatorStrategy } from "./MassIndexIndicatorStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let MassIndexIndicator = /*@__PURE__*/ (() => {
    class MassIndexIndicator extends StrategyBasedIndicator {
        bj() {
            return new MassIndexIndicator();
        }
        get_abv() {
            return new MassIndexIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return MassIndexIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
    }
    MassIndexIndicator.$t = /*@__PURE__*/ markType(MassIndexIndicator, 'MassIndexIndicator', StrategyBasedIndicator.$);
    return MassIndexIndicator;
})();
