// igniteui-react-charts/esm2015/lib/MedianPriceIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ItemwiseStrategyBasedIndicator } from "./ItemwiseStrategyBasedIndicator";
import { MedianPriceIndicatorStrategy } from "./MedianPriceIndicatorStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let MedianPriceIndicator = /*@__PURE__*/ (() => {
    class MedianPriceIndicator extends ItemwiseStrategyBasedIndicator {
        bj() {
            return new MedianPriceIndicator();
        }
        get_ach() {
            return new MedianPriceIndicatorStrategy();
        }
        get ach() {
            return this.get_ach();
        }
        get_ab8() {
            return MedianPriceIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
    }
    MedianPriceIndicator.$t = /*@__PURE__*/ markType(MedianPriceIndicator, 'MedianPriceIndicator', ItemwiseStrategyBasedIndicator.$);
    return MedianPriceIndicator;
})();
