// igniteui-react-charts/esm2015/lib/MoneyFlowIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { MoneyFlowIndexIndicatorStrategy } from "./MoneyFlowIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let MoneyFlowIndexIndicator = /*@__PURE__*/ (() => {
    class MoneyFlowIndexIndicator extends StrategyBasedIndicator {
        bj() {
            return new MoneyFlowIndexIndicator();
        }
        get_aa2() {
            return 1;
        }
        constructor() {
            super();
        }
        get_abv() {
            return new MoneyFlowIndexIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return MoneyFlowIndexIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(MoneyFlowIndexIndicator.ach));
        }
        set acg(a) {
            this.h(MoneyFlowIndexIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    MoneyFlowIndexIndicator.$t = /*@__PURE__*/ markType(MoneyFlowIndexIndicator, 'MoneyFlowIndexIndicator', StrategyBasedIndicator.$);
    MoneyFlowIndexIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(14, MoneyFlowIndexIndicator.$);
    return MoneyFlowIndexIndicator;
})();
