// igniteui-react-charts/esm2015/lib/MouseButtonMessage.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { InteractionMessage } from "./InteractionMessage";
import { MouseButtonAction_$type } from "./MouseButtonAction";
import { MouseButtonType_$type } from "./MouseButtonType";
import { EnumUtil, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let MouseButtonMessage = /*@__PURE__*/ (() => {
    class MouseButtonMessage extends InteractionMessage {
        constructor() {
            super(...arguments);
            this.b = 0;
            this.d = 0;
            this.g = null;
            this.f = 0;
        }
        get c() {
            return this.b;
        }
        set c(a) {
            this.b = a;
        }
        get e() {
            return this.d;
        }
        set e(a) {
            this.d = a;
        }
        get h() {
            return this.g;
        }
        set h(a) {
            this.g = a;
        }
        toString() {
            return "MouseButtonMessage[" + EnumUtil.getName(MouseButtonAction_$type, this.c) + ", " + EnumUtil.getName(MouseButtonType_$type, this.e) + ", " + this.h.toString() + "]";
        }
    }
    MouseButtonMessage.$t = /*@__PURE__*/ markType(MouseButtonMessage, 'MouseButtonMessage', InteractionMessage.$);
    return MouseButtonMessage;
})();
