// igniteui-react-charts/esm2015/lib/MovingAverageConvergenceDivergenceIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { MovingAverageConvergenceDivergenceIndicatorStrategy } from "./MovingAverageConvergenceDivergenceIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
/**
 * @hidden
 */
export let MovingAverageConvergenceDivergenceIndicator = /*@__PURE__*/ (() => {
    class MovingAverageConvergenceDivergenceIndicator extends StrategyBasedIndicator {
        bj() {
            return new MovingAverageConvergenceDivergenceIndicator();
        }
        get_aa2() {
            return 1;
        }
        constructor() {
            super();
            this.trendLineBrush = ((() => {
                let $ret = new Brush();
                $ret.fill = "#565656";
                return $ret;
            })());
            this.trendLineType = 10;
            this.trendLinePeriod = 9;
        }
        get_abv() {
            return new MovingAverageConvergenceDivergenceIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return MovingAverageConvergenceDivergenceIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get shortPeriod() {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.acj));
        }
        set shortPeriod(a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.acj, a);
        }
        ab3() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.aci));
        }
        set longPeriod(a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.aci, a);
        }
        ab0() {
            return this.longPeriod;
        }
        get signalPeriod() {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.ack));
        }
        set signalPeriod(a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.ack, a);
        }
        abb() {
            return this.signalPeriod;
        }
    }
    MovingAverageConvergenceDivergenceIndicator.$t = /*@__PURE__*/ markType(MovingAverageConvergenceDivergenceIndicator, 'MovingAverageConvergenceDivergenceIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    MovingAverageConvergenceDivergenceIndicator.acj = /*@__PURE__*/ StrategyBasedIndicator.acf(10, MovingAverageConvergenceDivergenceIndicator.$);
    MovingAverageConvergenceDivergenceIndicator.aci = /*@__PURE__*/ StrategyBasedIndicator.acb(30, MovingAverageConvergenceDivergenceIndicator.$);
    MovingAverageConvergenceDivergenceIndicator.ack = /*@__PURE__*/ StrategyBasedIndicator.ace(9, MovingAverageConvergenceDivergenceIndicator.$, "SignalPeriod");
    return MovingAverageConvergenceDivergenceIndicator;
})();
