// igniteui-react-charts/esm2015/lib/NegativeVolumeIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { NegativeVolumeIndexIndicatorStrategy } from "./NegativeVolumeIndexIndicatorStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let NegativeVolumeIndexIndicator = /*@__PURE__*/ (() => {
    class NegativeVolumeIndexIndicator extends StrategyBasedIndicator {
        bj() {
            return new NegativeVolumeIndexIndicator();
        }
        get_abv() {
            return new NegativeVolumeIndexIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return NegativeVolumeIndexIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
    }
    NegativeVolumeIndexIndicator.$t = /*@__PURE__*/ markType(NegativeVolumeIndexIndicator, 'NegativeVolumeIndexIndicator', StrategyBasedIndicator.$);
    return NegativeVolumeIndexIndicator;
})();
