// igniteui-react-charts/esm2015/lib/NormalRangeMessage_Spark.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RenderingMessage_Spark } from "./RenderingMessage_Spark";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let NormalRangeMessage_Spark = /*@__PURE__*/ (() => {
    class NormalRangeMessage_Spark extends RenderingMessage_Spark {
        constructor() {
            super(...arguments);
            this.g = 0;
            this.h = 0;
            this.f = 0;
            this.e = 0;
            this.c = false;
        }
        get k() {
            return this.g;
        }
        set k(a) {
            this.g = a;
        }
        get l() {
            return this.h;
        }
        set l(a) {
            this.h = a;
        }
        get j() {
            return this.f;
        }
        set j(a) {
            this.f = a;
        }
        get i() {
            return this.e;
        }
        set i(a) {
            this.e = a;
        }
        get d() {
            return this.c;
        }
        set d(a) {
            this.c = a;
        }
    }
    NormalRangeMessage_Spark.$t = /*@__PURE__*/ markType(NormalRangeMessage_Spark, 'NormalRangeMessage_Spark', RenderingMessage_Spark.$);
    return NormalRangeMessage_Spark;
})();
