// igniteui-react-charts/esm2015/lib/NumericYAxis_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StraightNumericAxisBase } from "./StraightNumericAxisBase";
import { IScaler_$type } from "./IScaler";
import { markType } from "igniteui-react-core";
import { VerticalAxisLabelPanel } from "./VerticalAxisLabelPanel";
import { Rect } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { RangeInfo } from "./RangeInfo";
import { VerticalLogarithmicScaler } from "./VerticalLogarithmicScaler";
import { FastVerticalLinearScaler } from "./FastVerticalLinearScaler";
import { isNaN_ } from "igniteui-react-core";
import { StraightNumericAxisBaseView } from "./StraightNumericAxisBaseView";
import { AxisDefaults } from "./AxisDefaults";
/**
 * @hidden
 */
export let NumericYAxis = /*@__PURE__*/ (() => {
    class NumericYAxis extends StraightNumericAxisBase {
        a4() {
            return new NumericYAxisView(this);
        }
        constructor() {
            super();
            this.ab = NumericYAxis.$;
            this.k4 = this.k2();
        }
        av() {
            return new VerticalAxisLabelPanel();
        }
        get_cp() {
            return true;
        }
        ek() {
            return this.kb().height;
        }
        el(a, b) {
            if (this.n2 != null) {
                return this.n2.p(a, b);
            }
            return this.n1.p(a, b);
        }
        hq(a, b, c, d) {
            if (this.n2 != null) {
                this.n2.v(a, b, c, d);
                return;
            }
            this.n1.v(a, b, c, d);
        }
        hp(a, b, c, d, e) {
            if (this.n2 != null) {
                this.n2.u(a, b, c, d, e);
                return;
            }
            this.n1.u(a, b, c, d, e);
        }
        eo(a, b) {
            if (this.n2 != null) {
                return this.n2.q(a, b);
            }
            return this.n1.q(a, b);
        }
        hr(a, b, c, d) {
            if (this.n2 != null) {
                this.n2.j(a, b, c, d);
                return;
            }
            this.n1.j(a, b, c, d);
        }
        oj(a, b) {
            if (this.bi == null || isNaN_(a) || isNaN_(b)) {
                return;
            }
            let c = this.n7(a, this);
            let d = this.n7(b, this);
            if (isNaN_(c) || isNaN_(d)) {
                return;
            }
            this.bi.wl = new Rect(0, this.bi.wl.x, d, this.bi.wl.width, c - d);
        }
        k2() {
            let a = super.k2();
            a.b.c = (b) => {
                if (this.az == null || this.az.visibility == 0) {
                    this.aw.bg = b;
                    if (this.az != null && (this.az.location == 7 || this.az.location == 6)) {
                        this.bi.n3();
                    }
                }
            };
            a.k = (b, c, d) => this.ht(c, d, b.aa, b.b, false);
            a.l = (b, c, d, e) => this.hu(c, d, e, b.aa);
            a.m = (b, c) => {
                let d = b.z;
                let e = new ScalerParams(0, b.ab, b.aa, this.ci, d);
                return this.el(c, e);
            };
            a.q = (b, c) => true;
            a.j = (b) => this.iq(b.u, b.n, b.aa, b.a, true);
            a.e = (b) => {
                b.n = this.az != null && (this.az.location == 7 || this.az.location == 3) ? b.aa.right : b.aa.left;
                let c = b.z;
                let d = new ScalerParams(0, b.ab, b.aa, this.ci, c);
                if (this.aj != null && this.aj.bi != null) {
                    let e = b.n;
                    let f = b.r;
                    let g = this.ii(e, f, b.ab, b.aa, b.z, b.aa.left, b.aa.right);
                    e = g.p0;
                    f = g.p1;
                    b.n = e;
                    b.r = f;
                }
            };
            a.p = (b, c, d) => {
                let e = Math.round(c);
                return e >= Math.floor(b.aa.top) && e <= Math.ceil(b.aa.bottom);
            };
            return a;
        }
        k5(a, b, c, d) {
            let e = d;
            let f = super.k5(a, b, c, d);
            let g = new ScalerParams(0, b, a, this.ch, c);
            let h = NaN;
            let i = NaN;
            if (!this.ch && b.top == 0) {
                i = this.l3;
            }
            else if (this.ch && b.bottom == 1) {
                h = this.l3;
            }
            if (isNaN_(h)) {
                h = this.eo(e.bottom, g);
            }
            if (isNaN_(i)) {
                i = this.eo(e.top, g);
            }
            let j = Math.min(h, i);
            let k = Math.max(h, i);
            this.l7 = j;
            this.l6 = k;
            let l = ((() => {
                let $ret = new RangeInfo();
                $ret.d = j;
                $ret.c = k;
                $ret.b = e.height;
                return $ret;
            })());
            f.k.add(l);
            return f;
        }
        ic(a) {
            super.ic(a);
            let b = this.bi != null ? this.bi.v4 : Rect.empty;
            let c = this.kh;
            let d = this.kc();
            let e = this.kb();
            let f = this.k4.c(a, c, b, d, e);
            if (f != null) {
                this.l2 = f.f.m;
                this.l5 = f.f.l;
            }
        }
        is(a, b) {
            super.is(a, b);
            if (b.height != a.height) {
                this.c1();
            }
        }
        get_a2() {
            return 1;
        }
        get a2() {
            return this.get_a2();
        }
        k7() {
            if (this.lo) {
                return new VerticalLogarithmicScaler();
            }
            switch (this.n0) {
                case 0: return new FastVerticalLinearScaler();
                case 1: return new VerticalLogarithmicScaler();
            }
            return null;
        }
        j$d(a, b) {
            return this.eo.apply(this, arguments);
        }
        j$c(a, b) {
            return this.el.apply(this, arguments);
        }
        j$e(a, b, c, d, e) {
            this.hp.apply(this, arguments);
        }
        j$f(a, b, c, d) {
            this.hq.apply(this, arguments);
        }
        j$g(a, b, c, d) {
            this.hr.apply(this, arguments);
        }
        get j$b() {
            return this.ch;
        }
        j$a() {
            return this.ag.apply(this, arguments);
        }
    }
    NumericYAxis.$t = /*@__PURE__*/ markType(NumericYAxis, 'NumericYAxis', StraightNumericAxisBase.$, [IScaler_$type]);
    return NumericYAxis;
})();
/**
 * @hidden
 */
export let NumericYAxisView = /*@__PURE__*/ (() => {
    class NumericYAxisView extends StraightNumericAxisBaseView {
        constructor(a) {
            super(a);
            this.bo = null;
            this.bo = a;
        }
        al() {
            super.al();
            this.a.jz = AxisDefaults.h;
        }
    }
    NumericYAxisView.$t = /*@__PURE__*/ markType(NumericYAxisView, 'NumericYAxisView', StraightNumericAxisBaseView.$);
    return NumericYAxisView;
})();
