// igniteui-react-charts/esm2015/lib/OverlayProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType, getInstanceType } from "igniteui-react-core";
import { IOverlayProxy_$type } from "./IOverlayProxy";
/**
 * @hidden
 */
export let OverlayProxy = /*@__PURE__*/ (() => {
    class OverlayProxy extends Base {
        matchesType(a) {
            if (a.e8) {
                return true;
            }
            return false;
        }
        setCloseMemberPath(a, b) {
            a.za = b;
        }
        setHighMemberPath(a, b) {
            a.zn = b;
        }
        setLowMemberPath(a, b) {
            a.zs = b;
        }
        setMultiplier(a, b) {
            if (getInstanceType(a).typeName == "BollingerBandsOverlay") {
                let s_ = a;
                let p_ = b;
                s_['multiplier'] = p_;
            }
        }
        setOpenMemberPath(a, b) {
            a.zw = b;
        }
        setVolumeMemberPath(a, b) {
            a.z0 = b;
        }
        setXAxis(a, b) {
            a.w6 = b;
        }
        setYAxis(a, b) {
            a.xp = b;
        }
    }
    OverlayProxy.$t = /*@__PURE__*/ markType(OverlayProxy, 'OverlayProxy', Base.$, [IOverlayProxy_$type]);
    return OverlayProxy;
})();
