// igniteui-react-charts/esm2015/lib/PathRenderingInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PathRenderingInfo = /*@__PURE__*/ (() => {
    class PathRenderingInfo extends Base {
        constructor() {
            super(...arguments);
            this.b = 0;
            this.e = null;
            this.d = null;
            this.f = null;
            this.a = false;
        }
        get c() {
            return this.b;
        }
        set c(a) {
            if (a < 0) {
                this.b = a * -1;
            }
            else {
                this.b = a;
            }
        }
    }
    PathRenderingInfo.$t = /*@__PURE__*/ markType(PathRenderingInfo, 'PathRenderingInfo');
    return PathRenderingInfo;
})();
