// igniteui-react-charts/esm2015/lib/PercentagePriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentagePriceOscillatorIndicatorStrategy } from "./PercentagePriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PercentagePriceOscillatorIndicator = /*@__PURE__*/ (() => {
    class PercentagePriceOscillatorIndicator extends StrategyBasedIndicator {
        bj() {
            return new PercentagePriceOscillatorIndicator();
        }
        get_abv() {
            return new PercentagePriceOscillatorIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return PercentagePriceOscillatorIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get shortPeriod() {
            return typeGetValue(this.c(PercentagePriceOscillatorIndicator.ach));
        }
        set shortPeriod(a) {
            this.h(PercentagePriceOscillatorIndicator.ach, a);
        }
        ab3() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(PercentagePriceOscillatorIndicator.acg));
        }
        set longPeriod(a) {
            this.h(PercentagePriceOscillatorIndicator.acg, a);
        }
        ab0() {
            return this.longPeriod;
        }
    }
    PercentagePriceOscillatorIndicator.$t = /*@__PURE__*/ markType(PercentagePriceOscillatorIndicator, 'PercentagePriceOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    PercentagePriceOscillatorIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acf(10, PercentagePriceOscillatorIndicator.$);
    PercentagePriceOscillatorIndicator.acg = /*@__PURE__*/ StrategyBasedIndicator.acb(30, PercentagePriceOscillatorIndicator.$);
    return PercentagePriceOscillatorIndicator;
})();
