// igniteui-react-charts/esm2015/lib/PercentageVolumeOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentageVolumeOscillatorIndicatorStrategy } from "./PercentageVolumeOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PercentageVolumeOscillatorIndicator = /*@__PURE__*/ (() => {
    class PercentageVolumeOscillatorIndicator extends StrategyBasedIndicator {
        bj() {
            return new PercentageVolumeOscillatorIndicator();
        }
        get_abv() {
            return new PercentageVolumeOscillatorIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return PercentageVolumeOscillatorIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get shortPeriod() {
            return typeGetValue(this.c(PercentageVolumeOscillatorIndicator.ach));
        }
        set shortPeriod(a) {
            this.h(PercentageVolumeOscillatorIndicator.ach, a);
        }
        ab3() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(PercentageVolumeOscillatorIndicator.acg));
        }
        set longPeriod(a) {
            this.h(PercentageVolumeOscillatorIndicator.acg, a);
        }
        ab0() {
            return this.longPeriod;
        }
    }
    PercentageVolumeOscillatorIndicator.$t = /*@__PURE__*/ markType(PercentageVolumeOscillatorIndicator, 'PercentageVolumeOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    PercentageVolumeOscillatorIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acf(10, PercentageVolumeOscillatorIndicator.$);
    PercentageVolumeOscillatorIndicator.acg = /*@__PURE__*/ StrategyBasedIndicator.acb(30, PercentageVolumeOscillatorIndicator.$);
    return PercentageVolumeOscillatorIndicator;
})();
