// igniteui-react-charts/esm2015/lib/PieChartStyleInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PieChartStyleInfo = /*@__PURE__*/ (() => {
    class PieChartStyleInfo extends Base {
        constructor() {
            super(...arguments);
            this.a = null;
            this.b = null;
            this.f = null;
            this.g = null;
            this.c = null;
            this.h = null;
            this.i = null;
            this.e = 0;
            this.d = 0;
        }
    }
    PieChartStyleInfo.$t = /*@__PURE__*/ markType(PieChartStyleInfo, 'PieChartStyleInfo');
    return PieChartStyleInfo;
})();
