// igniteui-react-charts/esm2015/lib/PointerTooltipPointerLocation.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available positions to anchor the pointer to the PointerTooltip.
 */
export var PointerTooltipPointerLocation = /*@__PURE__*/ (function (PointerTooltipPointerLocation) {
    /**
     * The anchor is decided automatically.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["Auto"] = 0] = "Auto";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["TopLeft"] = 1] = "TopLeft";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["TopMiddle"] = 2] = "TopMiddle";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["TopRight"] = 3] = "TopRight";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["RightTop"] = 4] = "RightTop";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["RightMiddle"] = 5] = "RightMiddle";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["RightBottom"] = 6] = "RightBottom";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["BottomRight"] = 7] = "BottomRight";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["BottomMiddle"] = 8] = "BottomMiddle";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["BottomLeft"] = 9] = "BottomLeft";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["LeftBottom"] = 10] = "LeftBottom";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["LeftMiddle"] = 11] = "LeftMiddle";
    /**
     * The anchor is top left of the content.
     */
    PointerTooltipPointerLocation[PointerTooltipPointerLocation["LeftTop"] = 12] = "LeftTop";
    return PointerTooltipPointerLocation;
})({});
/**
 * @hidden
 */
export let PointerTooltipPointerLocation_$type = /*@__PURE__*/ markEnum('PointerTooltipPointerLocation', 'Auto,0|TopLeft,1|TopMiddle,2|TopRight,3|RightTop,4|RightMiddle,5|RightBottom,6|BottomRight,7|BottomMiddle,8|BottomLeft,9|LeftBottom,10|LeftMiddle,11|LeftTop,12');
