// igniteui-react-charts/esm2015/lib/PolarScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PolarBase } from "./PolarBase";
import { markType } from "igniteui-react-core";
import { PolarBaseView } from "./PolarBaseView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
export let PolarScatterSeries = /*@__PURE__*/ (() => {
    class PolarScatterSeries extends PolarBase {
        bj() {
            return new PolarScatterSeries();
        }
        get_ei() {
            return true;
        }
        cs() {
            return new PolarScatterSeriesView(this);
        }
        q1(a) {
            super.q1(a);
            this.acg = a;
        }
        jl() {
            return 0.7;
        }
        w1() {
            if (this.w0 == 0) {
                return 2;
            }
            return super.w1();
        }
        constructor() {
            super();
            this.ab = PolarScatterSeries.$;
        }
    }
    PolarScatterSeries.$t = /*@__PURE__*/ markType(PolarScatterSeries, 'PolarScatterSeries', PolarBase.$);
    return PolarScatterSeries;
})();
/**
 * @hidden
 */
export let PolarScatterSeriesView = /*@__PURE__*/ (() => {
    class PolarScatterSeriesView extends PolarBaseView {
        constructor(a) {
            super(a);
            this.c8 = null;
            this.c8 = a;
        }
        au() {
            super.au();
            if (!this.q) {
                this.c8.xa = Defaults.e;
            }
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.v1 = a;
            this.e.jm = 3;
            this.e.jn = 2;
            this.e.jo = 2;
            this.e.hc = false;
        }
    }
    PolarScatterSeriesView.$t = /*@__PURE__*/ markType(PolarScatterSeriesView, 'PolarScatterSeriesView', PolarBaseView.$);
    return PolarScatterSeriesView;
})();
