// igniteui-react-charts/esm2015/lib/PolarLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { UnknownValuePlotting_$type } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { runOn, enumGetBox, markType, markDep } from "igniteui-react-core";
import { Path } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let PolarLineSeries = /*@__PURE__*/ (() => {
    class PolarLineSeries extends PolarLineSeriesBase {
        bj() {
            return new PolarLineSeries();
        }
        get_fk() {
            return true;
        }
        cs() {
            return new PolarLineSeriesView(this);
        }
        q1(a) {
            super.q1(a);
            this.acu = a;
        }
        constructor() {
            super();
            this.ab = PolarLineSeries.$;
        }
        acs(a) {
            let b = a;
            b.da();
        }
        act(a, b) {
            let c = b;
            c.db(a.m, this.id);
            this.y5.az(this, this.zx(), runOn(this, this.yn));
            let d = false;
            let e = this.y5.b;
            if (e != null) {
                d = true;
            }
            if (d) {
                this.abc(-1, this.zj.count, b.q);
            }
            this.y5.ad(b.dd, true, false, true, false);
        }
        ack() {
            return this.acv;
        }
        q9(a, b, c, d) {
            super.q9(a, b, c, d);
            switch (b) {
                case PolarLineSeries.$$p[0]:
                    this.rr(false);
                    this.ql();
                    break;
            }
        }
    }
    PolarLineSeries.$t = /*@__PURE__*/ markType(PolarLineSeries, 'PolarLineSeries', PolarLineSeriesBase.$);
    PolarLineSeries.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, PolarLineSeries, 'raisePropertyChanged', ['UnknownValuePlotting:acv:acx', [UnknownValuePlotting_$type, /*@__PURE__*/ enumGetBox(UnknownValuePlotting_$type, 1)]]);
    return PolarLineSeries;
})();
/**
 * @hidden
 */
export let PolarLineSeriesView = /*@__PURE__*/ (() => {
    class PolarLineSeriesView extends PolarLineSeriesBaseView {
        constructor(a) {
            super(a);
            this.c9 = null;
            this.dd = new Path();
            this.dc = new Path();
            this.c9 = a;
        }
        au() {
            super.au();
            if (!this.q) {
                this.b4.xa = Defaults.e;
            }
        }
        da() {
            this.dd.an = null;
            this.dc.an = null;
            this.as();
        }
        db(a, b) {
            this.c9.ach.d(this.dd, a, b);
            this.as();
        }
        bd() {
            super.bd();
        }
        be() {
            super.be();
            let a = this.br();
            this.dc.an = this.dd.an;
            this.dc._stroke = a;
            this.dc.ad = this.e.ie + 3;
        }
        a8(a, b) {
            super.a8(a, b);
            if (a.d) {
                if (b) {
                    a.t(this.dc);
                }
                else {
                    a.t(this.dd);
                }
            }
        }
        an(a) {
            super.an(a);
            this.e.bc.exportPathData(a, this.dd, "polarShape", ["Main"]);
        }
    }
    PolarLineSeriesView.$t = /*@__PURE__*/ markType(PolarLineSeriesView, 'PolarLineSeriesView', PolarLineSeriesBaseView.$);
    return PolarLineSeriesView;
})();
