// igniteui-react-charts/esm2015/lib/PreparationParams.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PreparationParams = /*@__PURE__*/ (() => {
    class PreparationParams extends Base {
        constructor() {
            super(...arguments);
            this.k = 0;
            this.l = 0;
            this.j = 0;
            this.i = 0;
            this.p = null;
            this.o = null;
            this.n = null;
            this.b = null;
            this.d = null;
            this.c = null;
            this.a = null;
            this.e = false;
            this.f = false;
            this.m = null;
            this.h = 0;
            this.g = 0;
        }
    }
    PreparationParams.$t = /*@__PURE__*/ markType(PreparationParams, 'PreparationParams');
    return PreparationParams;
})();
