// igniteui-react-charts/esm2015/lib/PriceVolumeTrendIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { PriceVolumeTrendIndicatorStrategy } from "./PriceVolumeTrendIndicatorStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PriceVolumeTrendIndicator = /*@__PURE__*/ (() => {
    class PriceVolumeTrendIndicator extends StrategyBasedIndicator {
        bj() {
            return new PriceVolumeTrendIndicator();
        }
        get_abv() {
            return new PriceVolumeTrendIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return PriceVolumeTrendIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
    }
    PriceVolumeTrendIndicator.$t = /*@__PURE__*/ markType(PriceVolumeTrendIndicator, 'PriceVolumeTrendIndicator', StrategyBasedIndicator.$);
    return PriceVolumeTrendIndicator;
})();
