// igniteui-react-charts/esm2015/lib/PropertyChangedMessage_Spark.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ConfigurationMessage_Spark } from "./ConfigurationMessage_Spark";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PropertyChangedMessage_Spark = /*@__PURE__*/ (() => {
    class PropertyChangedMessage_Spark extends ConfigurationMessage_Spark {
        constructor() {
            super(...arguments);
            this.f = null;
            this.c = null;
            this.b = null;
        }
        get propertyName() {
            return this.f;
        }
        set propertyName(a) {
            this.f = a;
        }
        get oldValue() {
            return this.c;
        }
        set oldValue(a) {
            this.c = a;
        }
        get newValue() {
            return this.b;
        }
        set newValue(a) {
            this.b = a;
        }
        toString() {
            let a = "null";
            let b = "null";
            if (this.oldValue != null) {
                a = this.oldValue.toString();
            }
            if (this.newValue != null) {
                b = this.newValue.toString();
            }
            return "PropertyChangedMessage[" + this.propertyName + ", " + a + ", " + b + "]";
        }
    }
    PropertyChangedMessage_Spark.$t = /*@__PURE__*/ markType(PropertyChangedMessage_Spark, 'PropertyChangedMessage_Spark', ConfigurationMessage_Spark.$);
    return PropertyChangedMessage_Spark;
})();
