// igniteui-react-charts/esm2015/lib/PropertyChangedMessage.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ConfigurationMessage } from "./ConfigurationMessage";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let PropertyChangedMessage = /*@__PURE__*/ (() => {
    class PropertyChangedMessage extends ConfigurationMessage {
        constructor() {
            super(...arguments);
            this.f = null;
            this.c = null;
            this.b = null;
        }
        get g() {
            return this.f;
        }
        set g(a) {
            this.f = a;
        }
        get e() {
            return this.c;
        }
        set e(a) {
            this.c = a;
        }
        get d() {
            return this.b;
        }
        set d(a) {
            this.b = a;
        }
        toString() {
            let a = "null";
            let b = "null";
            if (this.e != null) {
                a = this.e.toString();
            }
            if (this.d != null) {
                b = this.d.toString();
            }
            return "PropertyChangedMessage[" + this.g.toString() + ", " + a + ", " + b + "]";
        }
    }
    PropertyChangedMessage.$t = /*@__PURE__*/ markType(PropertyChangedMessage, 'PropertyChangedMessage', ConfigurationMessage.$);
    return PropertyChangedMessage;
})();
