// igniteui-react-charts/esm2015/lib/RateOfChangeAndMomentumIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RateOfChangeAndMomentumIndicatorStrategy } from "./RateOfChangeAndMomentumIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let RateOfChangeAndMomentumIndicator = /*@__PURE__*/ (() => {
    class RateOfChangeAndMomentumIndicator extends StrategyBasedIndicator {
        bj() {
            return new RateOfChangeAndMomentumIndicator();
        }
        get_abv() {
            return new RateOfChangeAndMomentumIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return RateOfChangeAndMomentumIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(RateOfChangeAndMomentumIndicator.ach));
        }
        set acg(a) {
            this.h(RateOfChangeAndMomentumIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    RateOfChangeAndMomentumIndicator.$t = /*@__PURE__*/ markType(RateOfChangeAndMomentumIndicator, 'RateOfChangeAndMomentumIndicator', StrategyBasedIndicator.$);
    RateOfChangeAndMomentumIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(14, RateOfChangeAndMomentumIndicator.$);
    return RateOfChangeAndMomentumIndicator;
})();
