// igniteui-react-charts/esm2015/lib/RenderOuterLabelMessage.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RenderingMessage } from "./RenderingMessage";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let RenderOuterLabelMessage = /*@__PURE__*/ (() => {
    class RenderOuterLabelMessage extends RenderingMessage {
        constructor() {
            super(...arguments);
            this.i = null;
            this.k = null;
            this.g = 0;
            this.f = null;
        }
        get j() {
            return this.i;
        }
        set j(a) {
            this.i = a;
        }
        get l() {
            return this.k;
        }
        set l(a) {
            this.k = a;
        }
        get h() {
            return this.g;
        }
        set h(a) {
            this.g = a;
        }
        toString() {
            return "RenderOuterLabelMessage[" + this.j.toString() + ", " + this.l.toString() + ", " + this.h.toString() + "]";
        }
    }
    RenderOuterLabelMessage.$t = /*@__PURE__*/ markType(RenderOuterLabelMessage, 'RenderOuterLabelMessage', RenderingMessage.$);
    return RenderOuterLabelMessage;
})();
