// igniteui-react-charts/esm2015/lib/RingSeriesVisualData.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { StringBuilder } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
export let RingSeriesVisualData = /*@__PURE__*/ (() => {
    class RingSeriesVisualData extends Base {
        constructor() {
            super(...arguments);
            this._name = null;
            this._labelExtent = 0;
            this._labelsPosition = null;
            this._leaderLineType = null;
            this._leaderLineVisibility = false;
            this._leaderLineMargin = 0;
            this._rings = null;
        }
        get name() {
            return this._name;
        }
        set name(a) {
            this._name = a;
        }
        get labelExtent() {
            return this._labelExtent;
        }
        set labelExtent(a) {
            this._labelExtent = a;
        }
        get labelsPosition() {
            return this._labelsPosition;
        }
        set labelsPosition(a) {
            this._labelsPosition = a;
        }
        get leaderLineType() {
            return this._leaderLineType;
        }
        set leaderLineType(a) {
            this._leaderLineType = a;
        }
        get leaderLineVisibility() {
            return this._leaderLineVisibility;
        }
        set leaderLineVisibility(a) {
            this._leaderLineVisibility = a;
        }
        get leaderLineMargin() {
            return this._leaderLineMargin;
        }
        set leaderLineMargin(a) {
            this._leaderLineMargin = a;
        }
        get rings() {
            return this._rings;
        }
        set rings(a) {
            this._rings = a;
        }
        serialize() {
            let a = new StringBuilder(0);
            let b = true;
            a.u("{");
            if (this.name != null) {
                if (b) {
                    b = false;
                }
                else {
                    a.l(", ");
                }
                a.l("name: ");
                a.u("'" + this.name + "'");
            }
            if (!isNaN_(this.labelExtent)) {
                if (b) {
                    b = false;
                }
                else {
                    a.l(", ");
                }
                a.l("labelExtent: ");
                a.u(this.labelExtent.toString());
            }
            if (this.labelsPosition != null) {
                if (b) {
                    b = false;
                }
                else {
                    a.l(", ");
                }
                a.l("labelsPosition: ");
                a.u("'" + this.labelsPosition + "'");
            }
            if (this.leaderLineType != null) {
                if (b) {
                    b = false;
                }
                else {
                    a.l(", ");
                }
                a.l("leaderLineType: ");
                a.u("'" + this.leaderLineType + "'");
            }
            a.l(", leaderLineVisibility: ");
            a.u(this.leaderLineVisibility ? "true" : "false");
            if (!isNaN_(this.leaderLineMargin)) {
                if (b) {
                    b = false;
                }
                else {
                    a.l(", ");
                }
                a.l("leaderLineMargin: ");
                a.u(this.leaderLineMargin.toString());
            }
            a.u(", rings: [");
            for (let c = 0; c < this.rings.count; c++) {
                if (c != 0) {
                    a.l(", ");
                }
                a.l(this.rings._inner[c].serialize());
            }
            a.u("]}");
            return a.toString();
        }
    }
    RingSeriesVisualData.$t = /*@__PURE__*/ markType(RingSeriesVisualData, 'RingSeriesVisualData');
    return RingSeriesVisualData;
})();
