// igniteui-react-charts/esm2015/lib/RingSeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RingSeriesBase } from "./RingSeriesBase";
import { Ring } from "./Ring";
import { ArcItem } from "./ArcItem";
import { fromEnum, markType } from "igniteui-react-core";
import { RingCollection } from "./RingCollection";
/**
 * @hidden
 */
export let RingSeries = /*@__PURE__*/ (() => {
    class RingSeries extends RingSeriesBase {
        constructor() {
            super();
            this.eo = null;
            this.eo = ((() => {
                let $ret = new Ring();
                $ret.c = this;
                $ret.e = false;
                return $ret;
            })());
            let a = ((() => {
                let $ret = new ArcItem();
                $ret.k = this.ba;
                $ret.d = this.eo;
                $ret.n = this.cs;
                $ret.c = this.ap;
                $ret.j = this.a8;
                return $ret;
            })());
            this.eo.g.add(a);
            this.ab = RingSeries.$;
        }
        c4(a, b, c, d) {
            super.c4(a, b, c, d);
            if (b == RingSeriesBase.$$p[6]) {
                for (let e = 0; e < this.eo.ringControl.arcs.count; e++) {
                    this.eo.ringControl.arcs.item(e).ef = d;
                }
            }
            if (b == RingSeriesBase.$$p[7]) {
                for (let f = 0; f < this.eo.ringControl.arcs.count; f++) {
                    this.eo.ringControl.arcs.item(f).ag = d;
                }
            }
            if (b == RingSeriesBase.$$p[28]) {
                for (let g = 0; g < this.eo.ringControl.arcs.count; g++) {
                    this.eo.ringControl.arcs.item(g).fa = d;
                }
            }
            if (b == RingSeriesBase.$$p[29]) {
                for (let h = 0; h < this.eo.ringControl.arcs.count; h++) {
                    this.eo.ringControl.arcs.item(h).aj = d;
                }
            }
            if (b == RingSeriesBase.$$p[18]) {
                for (let i = 0; i < this.eo.ringControl.arcs.count; i++) {
                    this.eo.ringControl.arcs.item(i).ew = d;
                }
            }
            if (b == RingSeriesBase.$$p[19]) {
                for (let j = 0; j < this.eo.ringControl.arcs.count; j++) {
                    this.eo.ringControl.arcs.item(j).ah = d;
                }
            }
            if (b == RingSeriesBase.$$p[21]) {
                for (let k = 0; k < this.eo.ringControl.arcs.count; k++) {
                    this.eo.ringControl.arcs.item(k).e1 = d;
                }
            }
            if (b == RingSeriesBase.$$p[22]) {
                for (let l = 0; l < this.eo.ringControl.arcs.count; l++) {
                    this.eo.ringControl.arcs.item(l).ai = d;
                }
            }
            if (b == RingSeriesBase.$$p[2]) {
                for (let m = 0; m < this.eo.ringControl.arcs.count; m++) {
                    this.eo.ringControl.arcs.item(m).a5 = d;
                }
            }
            if (b == RingSeriesBase.$$p[3]) {
                for (let n = 0; n < this.eo.ringControl.arcs.count; n++) {
                    this.eo.ringControl.arcs.item(n).a6 = d;
                }
            }
            if (b == RingSeriesBase.$$p[20]) {
                for (let o = 0; o < this.eo.ringControl.arcs.count; o++) {
                    this.eo.ringControl.arcs.item(o).ez = this.b3;
                }
            }
            if (b == RingSeriesBase.$$p[0]) {
                this.c2();
            }
            if (b == RingSeriesBase.$$p[33]) {
                if (this.eo.g != null && this.eo.g.count > 0) {
                    this.eo.g._inner[0].k = this.ba;
                }
            }
            if (b == RingSeriesBase.$$p[36]) {
                if (this.eo.g != null && this.eo.g.count > 0) {
                    this.eo.g._inner[0].n = this.cs;
                    this.c3();
                }
            }
            if (this.eo != null && this.aw.e()) {
                this.eo.o();
                if (this.eo.f()) {
                    this.eo.c.aw.l();
                }
            }
        }
        au() {
            let a = new RingCollection();
            if (this.eo.g._inner[0].g.count > 0) {
                a.add(this.eo);
            }
            return a;
        }
        c3() {
            if (this.eo != null) {
                this.eo.g._inner[0].h = this.itemsSource;
                this.eo.g._inner[0].p(this.ba);
                if (this.ax != null) {
                    this.ax.bx();
                    this.ax.b2();
                }
            }
        }
        c2() {
            if (this.eo != null) {
                this.eo.g._inner[0].a = this.ak;
            }
        }
        c8() {
            if (this.eo != null) {
                this.width = this.eo.q.width;
                this.height = this.eo.q.height;
                this.aw.m(this.eo.p.x, this.eo.p.y);
            }
        }
        c7() {
            if (this.eo != null) {
                for (let a of fromEnum(this.eo.ringControl.arcs.a)) {
                    a.gy();
                }
            }
        }
        cy(a, b) {
            super.cy(a, b);
            if (this.eo != null) {
                this.eo.l(this.itemsSource, a, b);
            }
        }
        cx() {
            super.cx();
            if (this.eo != null) {
                this.eo.k(this.itemsSource);
            }
        }
        cz(a, b) {
            super.cz(a, b);
            if (this.eo != null) {
                this.eo.m(this.itemsSource, a, b);
            }
        }
        c0(a, b, c) {
            super.c0(a, b, c);
            if (this.eo != null) {
                this.eo.n(this.itemsSource, a, b, c);
            }
        }
    }
    RingSeries.$t = /*@__PURE__*/ markType(RingSeries, 'RingSeries', RingSeriesBase.$);
    return RingSeries;
})();
