// igniteui-react-charts/esm2015/lib/SafeEnumerable.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, IEnumerable$1_$type, fromEnum, toEnum, Number_$type, markType, getEnumerator } from "igniteui-react-core";
import { isInfinity, isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
export let SafeEnumerable = /*@__PURE__*/ (() => {
    class SafeEnumerable extends Base {
        constructor(a) {
            super();
            this.a = null;
            this.a = a;
        }
        c(a) {
            if (isInfinity(a) || isNaN_(a)) {
                return 0;
            }
            return a;
        }
        *_b() {
            if (this.a == null) {
                return undefined;
            }
            for (let a of fromEnum(this.a)) {
                yield this.c(a);
            }
        }
        b() {
            return toEnum(() => this._b());
        }
        getEnumerator() {
            return getEnumerator(this.b());
        }
        getEnumeratorObject() {
            return this.getEnumerator();
        }
    }
    SafeEnumerable.$t = /*@__PURE__*/ markType(SafeEnumerable, 'SafeEnumerable', Base.$, [/*@__PURE__*/ IEnumerable$1_$type.specialize(Number_$type)]);
    return SafeEnumerable;
})();
