// igniteui-react-charts/esm2015/lib/ScalerParams.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { Rect } from "igniteui-react-core";
/**
 * @hidden
 */
export let ScalerParams = /*@__PURE__*/ (() => {
    class ScalerParams extends Base {
        constructor(a, ..._rest) {
            super();
            this.e = null;
            this.d = null;
            this.c = null;
            this.a = false;
            this.b = 0;
            a = (a == void 0) ? 0 : a;
            switch (a) {
                case 0:
                    {
                        let c = _rest[0];
                        let d = _rest[1];
                        let e = _rest[2];
                        let f = _rest[3];
                        this.e = c;
                        this.d = d;
                        this.c = f;
                        this.a = e;
                        this.b = NaN;
                    }
                    break;
                case 1:
                    {
                        let c = _rest[0];
                        let d = _rest[1];
                        let e = _rest[2];
                        let f = [c, d, e, Rect.empty];
                        {
                            let h = f[0];
                            let i = f[1];
                            let j = f[2];
                            let k = f[3];
                            this.e = h;
                            this.d = i;
                            this.c = k;
                            this.a = j;
                            this.b = NaN;
                        }
                    }
                    break;
            }
        }
    }
    ScalerParams.$t = /*@__PURE__*/ markType(ScalerParams, 'ScalerParams');
    return ScalerParams;
})();
