// igniteui-react-charts/esm2015/lib/ScatterLineSeriesView.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterBaseView } from "./ScatterBaseView";
import { Path } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { markType } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
export let ScatterLineSeriesView = /*@__PURE__*/ (() => {
    class ScatterLineSeriesView extends ScatterBaseView {
        constructor(a) {
            super(a);
            this.dl = new Path();
            this.dk = new Path();
        }
        au() {
            super.au();
            if (!this.q) {
                this.b4.xa = Defaults.e;
            }
        }
        bd() {
            super.bd();
        }
        be() {
            super.be();
            let a = this.br();
            this.dk.an = this.dl.an;
            this.dk._stroke = a;
            this.dk.ad = this.e.ie + 3;
        }
        a8(a, b) {
            super.a8(a, b);
            if (a.d) {
                if (b) {
                    a.t(this.dk);
                }
                else {
                    a.t(this.dl);
                }
            }
        }
        c9(a) {
            super.c9(a);
            this.dl.an = null;
            this.dk.an = null;
        }
        an(a) {
            super.an(a);
            this.e.bc.exportPathData(a, this.dl, "shape", ["Main"]);
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.v1 = a;
            this.e.jm = 3;
            this.e.jn = 1;
            this.e.jo = 4;
            this.e.hc = false;
        }
    }
    ScatterLineSeriesView.$t = /*@__PURE__*/ markType(ScatterLineSeriesView, 'ScatterLineSeriesView', ScatterBaseView.$);
    return ScatterLineSeriesView;
})();
