// igniteui-react-charts/esm2015/lib/ScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterBase } from "./ScatterBase";
import { markType } from "igniteui-react-core";
import { ScatterFrame } from "./ScatterFrame";
import { ScatterBaseView } from "./ScatterBaseView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
export let ScatterSeries = /*@__PURE__*/ (() => {
    class ScatterSeries extends ScatterBase {
        bj() {
            return new ScatterSeries();
        }
        get_ei() {
            return true;
        }
        cs() {
            return new ScatterSeriesView(this);
        }
        jl() {
            return 0.7;
        }
        w1() {
            if (this.w0 == 0) {
                return 2;
            }
            return super.w1();
        }
        q1(a) {
            super.q1(a);
            this.acn = a;
        }
        constructor() {
            super();
            this.ab = ScatterSeries.$;
            this.y2 = new ScatterFrame();
            this.y4 = new ScatterFrame();
            this.y0 = new ScatterFrame();
        }
    }
    ScatterSeries.$t = /*@__PURE__*/ markType(ScatterSeries, 'ScatterSeries', ScatterBase.$);
    return ScatterSeries;
})();
/**
 * @hidden
 */
export let ScatterSeriesView = /*@__PURE__*/ (() => {
    class ScatterSeriesView extends ScatterBaseView {
        constructor(a) {
            super(a);
            this.dk = null;
            this.dk = a;
        }
        au() {
            super.au();
            if (!this.q) {
                this.b4.xa = Defaults.e;
            }
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.v1 = a;
            this.e.jm = 3;
            this.e.jn = 2;
            this.e.jo = 2;
            this.e.hc = false;
        }
    }
    ScatterSeriesView.$t = /*@__PURE__*/ markType(ScatterSeriesView, 'ScatterSeriesView', ScatterBaseView.$);
    return ScatterSeriesView;
})();
