// igniteui-react-charts/esm2015/lib/ScatterPolylineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ShapeSeriesBase } from "./ShapeSeriesBase";
import { StyleSelector } from "./StyleSelector";
import { DependencyProperty } from "igniteui-react-core";
import { Style } from "igniteui-react-core";
import { markType, markDep, typeCast } from "igniteui-react-core";
import { ShapeSeriesViewBase } from "./ShapeSeriesViewBase";
import { ChartSelection } from "./ChartSelection";
import { List$1 } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { Path } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
/**
 * @hidden
 */
export let ScatterPolylineSeries = /*@__PURE__*/ (() => {
    class ScatterPolylineSeries extends ShapeSeriesBase {
        bj() {
            return new ScatterPolylineSeries();
        }
        get_fr() {
            return true;
        }
        constructor() {
            super();
            this.ab = ScatterPolylineSeries.$;
        }
        cs() {
            return new ScatterPolylineSeriesView(this);
        }
        get_x0() {
            return false;
        }
        q1(a) {
            super.q1(a);
            this.z7 = a;
        }
        get_x6() {
            return false;
        }
        q9(a, b, c, d) {
            super.q9(a, b, c, d);
            switch (b) {
                case ScatterPolylineSeries.$$p[0]:
                    this.z7.cy(this.aad);
                    this.rr(false);
                    this.ql();
                    break;
                case ScatterPolylineSeries.$$p[1]:
                    this.z7.cz(this.z8);
                    this.rr(false);
                    this.ql();
                    break;
            }
        }
        zn(a) {
            let b = a;
            b.cy(this.aad);
            b.cz(this.z8);
            super.zn(a);
        }
        a2(a, b) {
            let c = this.z7.c3(a, b, true);
            if (c != null) {
                return c.dataContext;
            }
            return super.a2(a, b);
        }
        ak(a, b) {
            let c = new List$1(ChartSelection.$, 0);
            this.p4(a, c, b);
            return c.toArray();
        }
        p4(a, b, c) {
            this.zf(a, b, c);
        }
    }
    ScatterPolylineSeries.$t = /*@__PURE__*/ markType(ScatterPolylineSeries, 'ScatterPolylineSeries', ShapeSeriesBase.$);
    ScatterPolylineSeries.$$p = /*@__PURE__*/ markDep(DependencyProperty, PropertyMetadata, ScatterPolylineSeries, 'raisePropertyChanged', ['ShapeStyle:aad:aab', [Style.$, null], 'ShapeStyleSelector:z8:aac', [StyleSelector.$, null]]);
    return ScatterPolylineSeries;
})();
/**
 * @hidden
 */
export let ScatterPolylineSeriesView = /*@__PURE__*/ (() => {
    class ScatterPolylineSeriesView extends ShapeSeriesViewBase {
        constructor(a) {
            super(a);
        }
        au() {
            super.au();
            if (!this.q) {
                this.b6.jp = Defaults.aa;
            }
        }
        c5(a, b) {
            return this.c6(a, b, false);
        }
        cm(a, b, c) {
            super.cm(a, b, c);
            ShapeSeriesViewBase.cl(this, a, b);
        }
        ck(a, b) {
            let c = typeCast(Path.$, a);
            if (c == null) {
                return;
            }
            c.an = b;
        }
        ac(a) {
            let b = super.ac(a);
            b._fill = null;
            return b;
        }
        ab(a) {
            let b = super.ab(a);
            b._fill = null;
            return b;
        }
        cb(a) {
            return a.width >= this.b6.yn || a.height >= this.b6.yn;
        }
        getDefaultTooltipTemplate() {
            let a = "<div class='ui-chart-default-tooltip-content'><span";
            let b = this.e.ba.a$u(this.e);
            if (!stringIsNullOrEmpty(b)) {
                a += " style='color:" + b + "'";
            }
            a += ">" + this.b6.title + "</span><br/>";
            a += "</div>";
            return a;
        }
    }
    ScatterPolylineSeriesView.$t = /*@__PURE__*/ markType(ScatterPolylineSeriesView, 'ScatterPolylineSeriesView', ShapeSeriesViewBase.$);
    return ScatterPolylineSeriesView;
})();
