// igniteui-react-charts/esm2015/lib/ScatterItemSearchMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available modes for searching for items in the scatter series.
 */
export var ScatterItemSearchMode = /*@__PURE__*/ (function (ScatterItemSearchMode) {
    /**
     * Search mode is automatically determined by the component.
     */
    ScatterItemSearchMode[ScatterItemSearchMode["Auto"] = 0] = "Auto";
    /**
     * When searching for a point, the closest visible point is selected.
     */
    ScatterItemSearchMode[ScatterItemSearchMode["ClosestVisiblePoint"] = 1] = "ClosestVisiblePoint";
    /**
     * When searching for a point, the closest point is selected, whether visible or not.
     * The search is skipped in performance critical scenarios if the number of points is over ItemSearchThreshold.
     */
    ScatterItemSearchMode[ScatterItemSearchMode["ClosestPoint"] = 2] = "ClosestPoint";
    /**
     * When searching for a point, the closest visible point along the closest visible line is used.
     * The search is skipped if the visible line points are over ItemSearchThreshold.
     */
    ScatterItemSearchMode[ScatterItemSearchMode["ClosestVisiblePointOnClosestLine"] = 3] = "ClosestVisiblePointOnClosestLine";
    /**
     * When searching for a point, the closest point along the closest line is used.
     * The search is skipped if the line points are over ItemSearchThreshold.
     */
    ScatterItemSearchMode[ScatterItemSearchMode["ClosestPointOnClosestLine"] = 4] = "ClosestPointOnClosestLine";
    /**
     * When searching for a point, the topmost visible point is selected.
     */
    ScatterItemSearchMode[ScatterItemSearchMode["TopVisiblePoint"] = 5] = "TopVisiblePoint";
    /**
     * No points will be searched for the in the scatter series.
     */
    ScatterItemSearchMode[ScatterItemSearchMode["None"] = 6] = "None";
    return ScatterItemSearchMode;
})({});
/**
 * @hidden
 */
export let ScatterItemSearchMode_$type = /*@__PURE__*/ markEnum('ScatterItemSearchMode', 'Auto,0|ClosestVisiblePoint,1|ClosestPoint,2|ClosestVisiblePointOnClosestLine,3|ClosestPointOnClosestLine,4|TopVisiblePoint,5|None,6');
