// igniteui-react-charts/esm2015/lib/ScatterAxisInfoCache.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ScatterAxisInfoCache = /*@__PURE__*/ (() => {
    class ScatterAxisInfoCache extends Base {
        constructor() {
            super(...arguments);
            this.a = null;
            this.b = null;
            this.d = false;
            this.e = false;
            this.c = null;
        }
    }
    ScatterAxisInfoCache.$t = /*@__PURE__*/ markType(ScatterAxisInfoCache, 'ScatterAxisInfoCache');
    return ScatterAxisInfoCache;
})();
