// igniteui-react-charts/esm2015/lib/SeriesRenderingArguments.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let SeriesRenderingArguments = /*@__PURE__*/ (() => {
    class SeriesRenderingArguments extends Base {
        constructor(a, b, c, d, e) {
            super();
            this.e = null;
            this.f = null;
            this.c = 0;
            this.a = false;
            this.d = null;
            this.b = false;
            this.c = a.kd;
            this.d = a;
            this.e = b;
            this.f = c;
            this.a = d;
            this.b = e;
        }
    }
    SeriesRenderingArguments.$t = /*@__PURE__*/ markType(SeriesRenderingArguments, 'SeriesRenderingArguments');
    return SeriesRenderingArguments;
})();
