// igniteui-react-charts/esm2015/lib/SeriesPointOfInterest.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let SeriesPointOfInterest = /*@__PURE__*/ (() => {
    class SeriesPointOfInterest extends Base {
        constructor() {
            super(...arguments);
            this.d = null;
            this.e = null;
            this.b = null;
            this.f = null;
            this.g = null;
            this.c = 0;
            this.a = null;
        }
    }
    SeriesPointOfInterest.$t = /*@__PURE__*/ markType(SeriesPointOfInterest, 'SeriesPointOfInterest');
    return SeriesPointOfInterest;
})();
