// igniteui-react-charts/esm2015/lib/SeriesSelectionMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes the Selection mode to use for the series in the chart.
 */
export var SeriesSelectionMode = /*@__PURE__*/ (function (SeriesSelectionMode) {
    /**
     * Selection mode is determined by the component.
     */
    SeriesSelectionMode[SeriesSelectionMode["Auto"] = 0] = "Auto";
    /**
     * No Selection is defined at the chart level.
     */
    SeriesSelectionMode[SeriesSelectionMode["None"] = 1] = "None";
    /**
     * Items will brighten when selected.
     */
    SeriesSelectionMode[SeriesSelectionMode["Brighten"] = 2] = "Brighten";
    /**
     * Items other than the selected items will fade.
     */
    SeriesSelectionMode[SeriesSelectionMode["FadeOthers"] = 3] = "FadeOthers";
    /**
     * Items other than the selected items will shift to grayscale.
     */
    SeriesSelectionMode[SeriesSelectionMode["GrayscaleOthers"] = 4] = "GrayscaleOthers";
    /**
     * Items selected will have the focus color as their thick outline.
     */
    SeriesSelectionMode[SeriesSelectionMode["FocusColorThickOutline"] = 5] = "FocusColorThickOutline";
    /**
     * Items selected will have the focus color as their outline.
     */
    SeriesSelectionMode[SeriesSelectionMode["FocusColorOutline"] = 6] = "FocusColorOutline";
    /**
     * Items selected will have the selection color as their thick outline.
     */
    SeriesSelectionMode[SeriesSelectionMode["SelectionColorThickOutline"] = 7] = "SelectionColorThickOutline";
    /**
     * Items selected will have the selection color as their outline.
     */
    SeriesSelectionMode[SeriesSelectionMode["SelectionColorOutline"] = 8] = "SelectionColorOutline";
    /**
     * Items selected will have the focus color as their fill.
     */
    SeriesSelectionMode[SeriesSelectionMode["FocusColorFill"] = 9] = "FocusColorFill";
    /**
     * Items selected will have the selection color as their fill.
     */
    SeriesSelectionMode[SeriesSelectionMode["SelectionColorFill"] = 10] = "SelectionColorFill";
    /**
     * Items selected will have a thicker outline than the rest.
     */
    SeriesSelectionMode[SeriesSelectionMode["ThickOutline"] = 11] = "ThickOutline";
    return SeriesSelectionMode;
})({});
/**
 * @hidden
 */
export let SeriesSelectionMode_$type = /*@__PURE__*/ markEnum('SeriesSelectionMode', 'Auto,0|None,1|Brighten,2|FadeOthers,3|GrayscaleOthers,4|FocusColorThickOutline,5|FocusColorOutline,6|SelectionColorThickOutline,7|SelectionColorOutline,8|FocusColorFill,9|SelectionColorFill,10|ThickOutline,11');
