// igniteui-react-charts/esm2015/lib/SeriesHighlightingBehavior.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes the highlighting mode to use for the series in the chart.
 */
export var SeriesHighlightingBehavior = /*@__PURE__*/ (function (SeriesHighlightingBehavior) {
    /**
     * Highlighting behavior is determined by the component.
     */
    SeriesHighlightingBehavior[SeriesHighlightingBehavior["Auto"] = 0] = "Auto";
    /**
     * Series are highlighted when the cursor is directly above them.
     */
    SeriesHighlightingBehavior[SeriesHighlightingBehavior["DirectlyOver"] = 1] = "DirectlyOver";
    /**
     * The nearest items to the cursor will be highlighted.
     */
    SeriesHighlightingBehavior[SeriesHighlightingBehavior["NearestItems"] = 2] = "NearestItems";
    /**
     * The nearest items to the cursor will be highlighted, the main shapes of the series will not be de-emphasized.
     */
    SeriesHighlightingBehavior[SeriesHighlightingBehavior["NearestItemsRetainMainShapes"] = 3] = "NearestItemsRetainMainShapes";
    /**
     * The nearest items to the cursor will be highlighted, the main shapes of the series closest to the cursor will not be de-emphasized.
     */
    SeriesHighlightingBehavior[SeriesHighlightingBehavior["NearestItemsAndSeries"] = 4] = "NearestItemsAndSeries";
    return SeriesHighlightingBehavior;
})({});
/**
 * @hidden
 */
export let SeriesHighlightingBehavior_$type = /*@__PURE__*/ markEnum('SeriesHighlightingBehavior', 'Auto,0|DirectlyOver,1|NearestItems,2|NearestItemsRetainMainShapes,3|NearestItemsAndSeries,4');
