// igniteui-react-charts/esm2015/lib/SetAreaSizeMessage.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RenderingMessage } from "./RenderingMessage";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let SetAreaSizeMessage = /*@__PURE__*/ (() => {
    class SetAreaSizeMessage extends RenderingMessage {
        constructor() {
            super(...arguments);
            this.f = false;
            this.j = 0;
            this.g = false;
            this.k = 0;
        }
        get h() {
            return this.f;
        }
        set h(a) {
            this.f = a;
        }
        get l() {
            return this.j;
        }
        set l(a) {
            this.j = a;
        }
        get i() {
            return this.g;
        }
        set i(a) {
            this.g = a;
        }
        get m() {
            return this.k;
        }
        set m(a) {
            this.k = a;
        }
        toString() {
            return "SetAreaSizeMessage[" + this.h.toString() + ", " + this.l.toString() + ", " + this.i.toString() + ", " + this.m.toString() + "]";
        }
    }
    SetAreaSizeMessage.$t = /*@__PURE__*/ markType(SetAreaSizeMessage, 'SetAreaSizeMessage', RenderingMessage.$);
    return SetAreaSizeMessage;
})();
