// igniteui-react-charts/esm2015/lib/ShapeItemSearchMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available modes for searching for items in the shape series.
 */
export var ShapeItemSearchMode = /*@__PURE__*/ (function (ShapeItemSearchMode) {
    /**
     * Search mode is automatically determined by the component.
     */
    ShapeItemSearchMode[ShapeItemSearchMode["Auto"] = 0] = "Auto";
    /**
     * When searching for a shape, the shape with the closest bounding box is selected.
     */
    ShapeItemSearchMode[ShapeItemSearchMode["ClosestBoundingBox"] = 1] = "ClosestBoundingBox";
    /**
     * When searching for a shape, the shape with a border closest to the target point is selected.
     * The search is skipped in performance critical scenarios if the number of bounds is over ItemSearchThreshold, or if the number of points in a shape is over ItemSearchPointsThreshold.
     */
    ShapeItemSearchMode[ShapeItemSearchMode["ClosestShape"] = 2] = "ClosestShape";
    /**
     * When searching for a shape, the closest point along the closest shape is used.
     * The search is skipped in performance critical scenarios if the number of bounds is over ItemSearchThreshold, or if the number of points in a shape is over ItemSearchPointsThreshold.
     */
    ShapeItemSearchMode[ShapeItemSearchMode["ClosestPointOnClosestShape"] = 3] = "ClosestPointOnClosestShape";
    /**
     * No shapes will be searched for the in the shape series.
     */
    ShapeItemSearchMode[ShapeItemSearchMode["None"] = 4] = "None";
    return ShapeItemSearchMode;
})({});
/**
 * @hidden
 */
export let ShapeItemSearchMode_$type = /*@__PURE__*/ markEnum('ShapeItemSearchMode', 'Auto,0|ClosestBoundingBox,1|ClosestShape,2|ClosestPointOnClosestShape,3|None,4');
