// igniteui-react-charts/esm2015/lib/SliceMouseMessage.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ModelUpdateMessage } from "./ModelUpdateMessage";
import { EnumUtil, markType } from "igniteui-react-core";
import { SliceMouseMessageType_$type } from "./SliceMouseMessageType";
/**
 * @hidden
 */
export let SliceMouseMessage = /*@__PURE__*/ (() => {
    class SliceMouseMessage extends ModelUpdateMessage {
        constructor() {
            super(...arguments);
            this.d = 0;
            this.f = null;
            this.b = 0;
            this.i = null;
            this.h = null;
        }
        get e() {
            return this.d;
        }
        set e(a) {
            this.d = a;
        }
        get item() {
            return this.f;
        }
        set item(a) {
            this.f = a;
        }
        get c() {
            return this.b;
        }
        set c(a) {
            this.b = a;
        }
        toString() {
            return "SliceMouseMessage[" + this.e.toString() + ", " + this.item.toString() + ", " + EnumUtil.getName(SliceMouseMessageType_$type, this.c) + "]";
        }
    }
    SliceMouseMessage.$t = /*@__PURE__*/ markType(SliceMouseMessage, 'SliceMouseMessage', ModelUpdateMessage.$);
    return SliceMouseMessage;
})();
