// igniteui-react-charts/esm2015/lib/SliceClickedMessage.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ModelUpdateMessage } from "./ModelUpdateMessage";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let SliceClickedMessage = /*@__PURE__*/ (() => {
    class SliceClickedMessage extends ModelUpdateMessage {
        constructor() {
            super(...arguments);
            this.b = 0;
            this.d = null;
            this.f = null;
        }
        get c() {
            return this.b;
        }
        set c(a) {
            this.b = a;
        }
        get item() {
            return this.d;
        }
        set item(a) {
            this.d = a;
        }
        toString() {
            return "SliceClickedMessage[" + this.c.toString() + ", " + this.item.toString() + "]";
        }
    }
    SliceClickedMessage.$t = /*@__PURE__*/ markType(SliceClickedMessage, 'SliceClickedMessage', ModelUpdateMessage.$);
    return SliceClickedMessage;
})();
