// igniteui-react-charts/esm2015/lib/SortedReadOnlyDoubleCollection.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, IList$1_$type, Number_$type, toEnum, markType } from "igniteui-react-core";
import { ReadOnlyCollection$1 } from "igniteui-react-core";
/**
 * @hidden
 */
export let SortedReadOnlyDoubleCollection = /*@__PURE__*/ (() => {
    class SortedReadOnlyDoubleCollection extends Base {
        constructor(a, b) {
            super();
            this.b = null;
            this.a = null;
            this.b = new ReadOnlyCollection$1(Number_$type, 1, a);
            this.a = b;
        }
        indexOf(a) {
            let b = this.b.indexOf(a);
            return this.a.indexOf(b);
        }
        insert(a, b) {
            this.b.insert(a, b);
        }
        removeAt(a) {
            this.b.removeAt(a);
        }
        item(a, b) {
            if (arguments.length === 2) {
                this.b.item(a, b);
                return b;
            }
            else {
                let b = this.a.item(a);
                return this.b.item(b);
            }
        }
        add(a) {
            this.b.add(a);
        }
        clear() {
            this.b.clear();
        }
        contains(a) {
            return this.b.contains(a);
        }
        copyTo(a, b) {
            for (let c = b; c < a.length; c++) {
                a[c] = this.item(c);
            }
        }
        get count() {
            return this.b.count;
        }
        get isReadOnly() {
            return this.b.isReadOnly;
        }
        remove(a) {
            return this.b.remove(a);
        }
        *_getEnumerator() {
            for (let a = 0; a < this.b.count; a++) {
                yield this.item(a);
            }
        }
        getEnumerator() {
            return toEnum(() => this._getEnumerator()).getEnumerator();
        }
        getEnumeratorObject() {
            return this.getEnumerator();
        }
    }
    SortedReadOnlyDoubleCollection.$t = /*@__PURE__*/ markType(SortedReadOnlyDoubleCollection, 'SortedReadOnlyDoubleCollection', Base.$, [/*@__PURE__*/ IList$1_$type.specialize(Number_$type)]);
    return SortedReadOnlyDoubleCollection;
})();
