// igniteui-react-charts/esm2015/lib/SparklineAxisVisualData.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { StringBuilder } from "igniteui-react-core";
/**
 * @hidden
 */
export let SparklineAxisVisualData = /*@__PURE__*/ (() => {
    class SparklineAxisVisualData extends Base {
        constructor() {
            super();
            this._name = null;
            this._minLabel = null;
            this._maxLabel = null;
            this._axisLine = null;
        }
        get name() {
            return this._name;
        }
        set name(a) {
            this._name = a;
        }
        get minLabel() {
            return this._minLabel;
        }
        set minLabel(a) {
            this._minLabel = a;
        }
        get maxLabel() {
            return this._maxLabel;
        }
        set maxLabel(a) {
            this._maxLabel = a;
        }
        get axisLine() {
            return this._axisLine;
        }
        set axisLine(a) {
            this._axisLine = a;
        }
        scaleByViewport(a) {
        }
        serialize() {
            let a = new StringBuilder(0);
            a.u("{");
            a.u("name: \"" + this.name + "\"");
            if (this.minLabel != null) {
                a.u(",");
                a.l("minLabel: " + this.minLabel.serialize());
            }
            if (this.maxLabel != null) {
                a.u(",");
                a.l("maxLabel: " + this.maxLabel.serialize());
            }
            if (this.axisLine != null) {
                a.u(",");
                a.l("axisLine: " + this.axisLine.serialize());
            }
            a.u("");
            a.u("}");
            return a.toString();
        }
    }
    SparklineAxisVisualData.$t = /*@__PURE__*/ markType(SparklineAxisVisualData, 'SparklineAxisVisualData');
    return SparklineAxisVisualData;
})();
