// igniteui-react-charts/esm2015/lib/SparkFrame.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Array_$type, Point_$type, markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
/**
 * @hidden
 */
export let SparkFrame = /*@__PURE__*/ (() => {
    class SparkFrame extends Base {
        constructor() {
            super(...arguments);
            this.a = new List$1(Array_$type, 0);
            this.f = new List$1(Point_$type, 0);
            this.d = new List$1(Point_$type, 0);
            this.e = new List$1(Point_$type, 0);
            this.c = new List$1(Point_$type, 0);
            this.b = new List$1(Point_$type, 0);
            this.m = null;
            this.n = null;
        }
        get g() {
            return this.a;
        }
        set g(a) {
            this.a = a;
        }
        get l() {
            return this.f;
        }
        set l(a) {
            this.f = a;
        }
        get j() {
            return this.d;
        }
        set j(a) {
            this.d = a;
        }
        get k() {
            return this.e;
        }
        set k(a) {
            this.e = a;
        }
        get i() {
            return this.c;
        }
        set i(a) {
            this.c = a;
        }
        get h() {
            return this.b;
        }
        set h(a) {
            this.b = a;
        }
    }
    SparkFrame.$t = /*@__PURE__*/ markType(SparkFrame, 'SparkFrame');
    return SparkFrame;
})();
