// igniteui-react-charts/esm2015/lib/SplineAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SplineSeriesBase } from "./SplineSeriesBase";
import { CategoryLineRasterizer } from "./CategoryLineRasterizer";
import { typeCast, runOn, markType } from "igniteui-react-core";
import { Path } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { SplineSeriesBaseView } from "./SplineSeriesBaseView";
import { Defaults } from "./Defaults";
import { ISortingAxis_$type } from "./ISortingAxis";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
export let SplineAreaSeries = /*@__PURE__*/ (() => {
    class SplineAreaSeries extends SplineSeriesBase {
        bj() {
            return new SplineAreaSeries();
        }
        get_eu() {
            return true;
        }
        get_f0() {
            return true;
        }
        cs() {
            return new SplineAreaSeriesView(this);
        }
        y4() {
            return 1;
        }
        q1(a) {
            super.q1(a);
            this.adh = a;
        }
        constructor() {
            super();
            this.adg = null;
            this.adg = new CategoryLineRasterizer();
            this.ab = SplineAreaSeries.$;
        }
        preferredCategoryMode(a) {
            return 0;
        }
        cy() {
            if (this.cz == 0) {
                return 2;
            }
            return super.cy();
        }
        o5(a, b) {
            super.o5(a, b);
            let c = b;
            c.di();
        }
        gw(a, b) {
            let c = this.wc(this.wh(a), true, false);
            if (this.acn(a, c, b)) {
                return true;
            }
            if (this.gv(a, b)) {
                return true;
            }
            if (this.xh(a, b)) {
                return true;
            }
            return false;
        }
        aak(a, b) {
            super.aak(a, b);
            let c = this.y3.cw.d;
            let d = typeCast(SplineAreaSeriesView.$, b);
            let e = a.f;
            let f = this.getEffectiveViewport1(b);
            this.y2.a2(this, this.zm(), runOn(this, this.ym), this.zv(b), this.zw(b));
            let g = false;
            let h = this.y2.b;
            if (h != null) {
                g = true;
            }
            if (g) {
                let i = new ScalerParams(0, b.by, b.bx, this.acd.ch, f);
                this.aag(e, -1, this.valueColumn.count, this.acd, i, b.q, -1);
            }
            let j = d.du;
            let k = d.dv;
            let l = d.dw;
            let m = d.dx;
            this.y2.ad(j, true, false, true, true);
            this.y2.ad(k, true, false, true, true);
            this.y2.ad(l, false, true, false, false);
            this.y2.ad(m, false, true, false, false);
            l._opacity = this.y2.j * this.h9;
            m._opacity = 0.5 * this.y2.j * this.h9;
            if (b.c0(a)) {
                d.dk(a.f.count, e, true, c, this.id, (n, o, p, q, r) => this.abt(n, a.f.count, b), 1, a.f.count == 1 ? null : this.aau(e, e.count - 1, b.bx, b.by, true));
                b.c3(a);
            }
            let n = this.fetchYAxis();
        }
        get useHighMarkerFidelity() {
            return this.zq;
        }
    }
    SplineAreaSeries.$t = /*@__PURE__*/ markType(SplineAreaSeries, 'SplineAreaSeries', SplineSeriesBase.$);
    return SplineAreaSeries;
})();
/**
 * @hidden
 */
export let SplineAreaSeriesView = /*@__PURE__*/ (() => {
    class SplineAreaSeriesView extends SplineSeriesBaseView {
        constructor(a) {
            super(a);
            this.dh = null;
            this.dq = new Path();
            this.ds = new Path();
            this.dr = new Path();
            this.dt = new Path();
            this.dl = new Path();
            this.dn = new Path();
            this.dm = new Path();
            this.dp = new Path();
            this.dh = a;
        }
        au() {
            super.au();
            if (!this.q) {
                this.dh.xa = Defaults.e;
            }
        }
        get du() {
            return this.ds;
        }
        get dv() {
            return this.dt;
        }
        get dw() {
            return this.dq;
        }
        get dx() {
            return this.dr;
        }
        di() {
            this.dq.an = null;
            this.dr.an = null;
            this.ds.an = null;
            this.dt.an = null;
            this.as();
        }
        dj(a, b, c, d, e, f, g) {
            this.dk(a, b, c, d, e, f, g, null);
        }
        dk(a, b, c, d, e, f, g, h) {
            this.dh.adg.a = typeCast(ISortingAxis_$type, this.cy.fetchXAxis()) !== null ? true : false;
            this.dh.adg.j(this.dq, this.ds, this.dr, this.dt, a, b, c, d, e, f, g, h);
            this.as();
        }
        be() {
            super.be();
            this.dl.an = this.dq.an;
            this.dn.an = this.ds.an;
            this.dm.an = this.dr.an;
            this.dp.an = this.dt.an;
            let a = this.br();
            this.dl._fill = a;
            this.dl._opacity = 1;
            this.dm._fill = a;
            this.dm._opacity = 1;
            this.dn._stroke = a;
            this.dn.ad = this.e.ie + 3;
            this.dp._stroke = a;
            this.dp.ad = this.e.ie + 3;
        }
        a8(a, b) {
            super.a8(a, b);
            if (b) {
                a.t(this.dl);
                a.t(this.dm);
                a.t(this.dn);
                a.t(this.dp);
            }
            else {
                a.t(this.dq);
                a.t(this.dr);
                a.t(this.ds);
                a.t(this.dt);
            }
        }
        an(a) {
            super.an(a);
            this.e.bc.exportPathData(a, this.ds, "lowerShape", ["Lower"]);
            this.e.bc.exportPathData(a, this.dt, "upperShape", ["Upper", "Main"]);
            this.e.bc.exportPathData(a, this.dr, "translucentShape", ["Translucent"]);
            this.e.bc.exportPathData(a, this.dq, "fillShape", ["Fill"]);
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.v1 = a;
            this.e.jm = 5;
            this.e.jn = 1;
            this.e.jo = -3;
        }
    }
    SplineAreaSeriesView.$t = /*@__PURE__*/ markType(SplineAreaSeriesView, 'SplineAreaSeriesView', SplineSeriesBaseView.$);
    return SplineAreaSeriesView;
})();
