// igniteui-react-charts/esm2015/lib/Stacked100SplineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedSplineSeries } from "./StackedSplineSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedSplineSeriesView } from "./StackedSplineSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let Stacked100SplineSeries = /*@__PURE__*/ (() => {
    class Stacked100SplineSeries extends StackedSplineSeries {
        cs() {
            return new Stacked100SplineSeriesView(this);
        }
        q1(a) {
            super.q1(a);
            this.ab6 = a;
        }
        get_aa4() {
            return true;
        }
        constructor() {
            super();
            this.ab = Stacked100SplineSeries.$;
        }
        ac() {
            super.ac();
            this.rr(false);
        }
        abl() {
            super.abl();
            Stacked100DataPreparer.a(this);
        }
        aba(a, b) {
            let c = 0;
            for (let d = 0; d < this.series.count; d++) {
                c += this.series._inner[d].ak.item(b);
            }
            return (this.series._inner[a].ak.item(b) / c) * 100;
        }
        get useHighMarkerFidelity() {
            return this.zq;
        }
    }
    Stacked100SplineSeries.$t = /*@__PURE__*/ markType(Stacked100SplineSeries, 'Stacked100SplineSeries', StackedSplineSeries.$, [IStacked100Series_$type]);
    return Stacked100SplineSeries;
})();
/**
 * @hidden
 */
export let Stacked100SplineSeriesView = /*@__PURE__*/ (() => {
    class Stacked100SplineSeriesView extends StackedSplineSeriesView {
        constructor(a) {
            super(a);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.xa = Defaults.e;
            }
        }
    }
    Stacked100SplineSeriesView.$t = /*@__PURE__*/ markType(Stacked100SplineSeriesView, 'Stacked100SplineSeriesView', StackedSplineSeriesView.$);
    return Stacked100SplineSeriesView;
})();
