// igniteui-react-charts/esm2015/lib/StackedLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalStackedSeriesBase } from "./HorizontalStackedSeriesBase";
import { markType } from "igniteui-react-core";
import { AxisRange } from "./AxisRange";
import { ChartSelection } from "./ChartSelection";
import { List$1 } from "igniteui-react-core";
import { StackedSeriesView } from "./StackedSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let StackedLineSeries = /*@__PURE__*/ (() => {
    class StackedLineSeries extends HorizontalStackedSeriesBase {
        bj() {
            return new StackedLineSeries();
        }
        get_fk() {
            return true;
        }
        constructor() {
            super();
            this.ab = StackedLineSeries.$;
        }
        ac() {
            super.ac();
            this.rr(false);
        }
        cs() {
            return new StackedLineSeriesView(this);
        }
        q1(a) {
            super.q1(a);
            this.ab2 = a;
        }
        ay(a) {
            if (this.itemsSource == null) {
                return null;
            }
            if (a == this.xAxis) {
                return new AxisRange(0, this.dn.count - 1);
            }
            if (a == this.yAxis) {
                this.abl();
                return new AxisRange(this.abc, this.abb);
            }
            return null;
        }
        preferredCategoryMode(a) {
            return 0;
        }
        ak(a, b) {
            let c = new List$1(ChartSelection.$, 0);
            this.p4(a, c, b);
            return c.toArray();
        }
        p4(a, b, c) {
            this.aac(a, b, c, this.yt, 0);
        }
        get useHighMarkerFidelity() {
            return this.zq;
        }
    }
    StackedLineSeries.$t = /*@__PURE__*/ markType(StackedLineSeries, 'StackedLineSeries', HorizontalStackedSeriesBase.$);
    return StackedLineSeries;
})();
/**
 * @hidden
 */
export let StackedLineSeriesView = /*@__PURE__*/ (() => {
    class StackedLineSeriesView extends StackedSeriesView {
        constructor(a) {
            super(a);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.xa = Defaults.e;
            }
        }
    }
    StackedLineSeriesView.$t = /*@__PURE__*/ markType(StackedLineSeriesView, 'StackedLineSeriesView', StackedSeriesView.$);
    return StackedLineSeriesView;
})();
