// igniteui-react-charts/esm2015/lib/Stacked100LineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedLineSeries } from "./StackedLineSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedLineSeriesView } from "./StackedLineSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let Stacked100LineSeries = /*@__PURE__*/ (() => {
    class Stacked100LineSeries extends StackedLineSeries {
        cs() {
            return new Stacked100LineSeriesView(this);
        }
        q1(a) {
            super.q1(a);
            this.ab3 = a;
        }
        get_aa4() {
            return true;
        }
        constructor() {
            super();
            this.ab = Stacked100LineSeries.$;
        }
        ac() {
            super.ac();
            this.rr(false);
        }
        abl() {
            super.abl();
            Stacked100DataPreparer.a(this);
        }
        aba(a, b) {
            let c = 0;
            for (let d = 0; d < this.series.count; d++) {
                c += this.series._inner[d].ak.item(b);
            }
            return (this.series._inner[a].ak.item(b) / c) * 100;
        }
        get useHighMarkerFidelity() {
            return this.zq;
        }
    }
    Stacked100LineSeries.$t = /*@__PURE__*/ markType(Stacked100LineSeries, 'Stacked100LineSeries', StackedLineSeries.$, [IStacked100Series_$type]);
    return Stacked100LineSeries;
})();
/**
 * @hidden
 */
export let Stacked100LineSeriesView = /*@__PURE__*/ (() => {
    class Stacked100LineSeriesView extends StackedLineSeriesView {
        constructor(a) {
            super(a);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.xa = Defaults.e;
            }
        }
    }
    Stacked100LineSeriesView.$t = /*@__PURE__*/ markType(Stacked100LineSeriesView, 'Stacked100LineSeriesView', StackedLineSeriesView.$);
    return Stacked100LineSeriesView;
})();
