// igniteui-react-charts/esm2015/lib/Stacked100ColumnSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedColumnSeries } from "./StackedColumnSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { Defaults } from "./Defaults";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { isNaN_ } from "igniteui-react-core";
import { StackedColumnSeriesView } from "./StackedColumnSeriesView";
import { Stacked100ColumnBucketCalculator } from "./Stacked100ColumnBucketCalculator";
/**
 * @hidden
 */
export let Stacked100ColumnSeries = /*@__PURE__*/ (() => {
    class Stacked100ColumnSeries extends StackedColumnSeries {
        get_aa4() {
            return true;
        }
        jk() {
            if (!isNaN_(this.jj)) {
                return super.jk();
            }
            return Defaults.h;
        }
        constructor() {
            super();
            this.ab = Stacked100ColumnSeries.$;
        }
        ac() {
            super.ac();
            this.rr(false);
        }
        cs() {
            return new Stacked100ColumnSeriesView(this);
        }
        q1(a) {
            super.q1(a);
            this.ab9 = a;
        }
        aav() {
            return this.ab9;
        }
        abl() {
            super.abl();
            Stacked100DataPreparer.a(this);
        }
        aba(a, b) {
            let c = 0;
            for (let d = 0; d < this.series.count; d++) {
                c += this.series._inner[d].ak.item(b);
            }
            return (this.series._inner[a].ak.item(b) / c) * 100;
        }
        get useHighMarkerFidelity() {
            return this.zq;
        }
    }
    Stacked100ColumnSeries.$t = /*@__PURE__*/ markType(Stacked100ColumnSeries, 'Stacked100ColumnSeries', StackedColumnSeries.$, [IStacked100Series_$type]);
    return Stacked100ColumnSeries;
})();
/**
 * @hidden
 */
export let Stacked100ColumnSeriesView = /*@__PURE__*/ (() => {
    class Stacked100ColumnSeriesView extends StackedColumnSeriesView {
        constructor(a) {
            super(a);
            this.da = null;
            this.da = a;
        }
        cx() {
            return new Stacked100ColumnBucketCalculator(this);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.xa = Defaults.c;
                this.c5.jp = Defaults.i;
            }
        }
    }
    Stacked100ColumnSeriesView.$t = /*@__PURE__*/ markType(Stacked100ColumnSeriesView, 'Stacked100ColumnSeriesView', StackedColumnSeriesView.$);
    return Stacked100ColumnSeriesView;
})();
