// igniteui-react-charts/esm2015/lib/StandardDeviationIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { StandardDeviationIndicatorStrategy } from "./StandardDeviationIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let StandardDeviationIndicator = /*@__PURE__*/ (() => {
    class StandardDeviationIndicator extends StrategyBasedIndicator {
        bj() {
            return new StandardDeviationIndicator();
        }
        get_abv() {
            return new StandardDeviationIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return StandardDeviationIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(StandardDeviationIndicator.ach));
        }
        set acg(a) {
            this.h(StandardDeviationIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    StandardDeviationIndicator.$t = /*@__PURE__*/ markType(StandardDeviationIndicator, 'StandardDeviationIndicator', StrategyBasedIndicator.$);
    StandardDeviationIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(20, StandardDeviationIndicator.$);
    return StandardDeviationIndicator;
})();
