// igniteui-react-charts/esm2015/lib/StochRSIIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { StochRSIIndicatorStrategy } from "./StochRSIIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let StochRSIIndicator = /*@__PURE__*/ (() => {
    class StochRSIIndicator extends StrategyBasedIndicator {
        bj() {
            return new StochRSIIndicator();
        }
        get_abv() {
            return new StochRSIIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return StochRSIIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(StochRSIIndicator.ach));
        }
        set acg(a) {
            this.h(StochRSIIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    StochRSIIndicator.$t = /*@__PURE__*/ markType(StochRSIIndicator, 'StochRSIIndicator', StrategyBasedIndicator.$);
    StochRSIIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(14, StochRSIIndicator.$);
    return StochRSIIndicator;
})();
