// igniteui-react-charts/esm2015/lib/SupportingCalculation$1.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, String_$type, markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
/**
 * @hidden
 */
export let SupportingCalculation$1 = /*@__PURE__*/ (() => {
    class SupportingCalculation$1 extends Base {
        constructor($tCalculationStrategy, a, ..._rest) {
            super();
            this.$tCalculationStrategy = null;
            this.a = null;
            this.d = null;
            this.$tCalculationStrategy = $tCalculationStrategy;
            this.$type = this.$type.specialize(this.$tCalculationStrategy);
            a = (a == void 0) ? 0 : a;
            switch (a) {
                case 0:
                    {
                        let c = _rest[0];
                        this.a = c;
                        this.d = new List$1(String_$type, 0);
                    }
                    break;
                case 1:
                    {
                        let c = _rest[0];
                        let d = _rest[1];
                        this.a = c;
                        this.d = new List$1(String_$type, 1, d);
                    }
                    break;
            }
        }
        get b() {
            return this.a;
        }
        get c() {
            return this.d;
        }
    }
    SupportingCalculation$1.$t = /*@__PURE__*/ markType(SupportingCalculation$1, 'SupportingCalculation$1');
    return SupportingCalculation$1;
})();
