// igniteui-react-charts/esm2015/lib/TickmarkValues.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DependencyObject } from "igniteui-react-core";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let TickmarkValues = /*@__PURE__*/ (() => {
    class TickmarkValues extends DependencyObject {
        constructor() {
            super(...arguments);
            this.o = 0;
            this.m = 0;
            this.n = 0;
            this.l = 0;
            this.p = 0;
            this.q = 0;
            this.r = 0;
        }
        s(a) {
            this.o = a.m;
        }
    }
    TickmarkValues.$t = /*@__PURE__*/ markType(TickmarkValues, 'TickmarkValues', DependencyObject.$);
    return TickmarkValues;
})();
