// igniteui-react-charts/esm2015/lib/TickmarkValuesInitializationParameters.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let TickmarkValuesInitializationParameters = /*@__PURE__*/ (() => {
    class TickmarkValuesInitializationParameters extends Base {
        constructor() {
            super(...arguments);
            this.n = 0;
            this.m = 0;
            this.g = 0;
            this.f = 0;
            this.k = 0;
            this.d = false;
            this.l = 0;
            this.i = 0;
            this.q = 0;
            this.a = 0;
            this.r = 0;
            this.s = null;
            this.t = null;
            this.e = false;
            this.o = null;
            this.b = null;
            this.c = null;
            this.j = 0;
            this.h = 0;
            this.p = 0;
        }
    }
    TickmarkValuesInitializationParameters.$t = /*@__PURE__*/ markType(TickmarkValuesInitializationParameters, 'TickmarkValuesInitializationParameters');
    return TickmarkValuesInitializationParameters;
})();
