// igniteui-react-charts/esm2015/lib/TimeAxisIntervalType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Enum representing a unit of time.
 */
export var TimeAxisIntervalType = /*@__PURE__*/ (function (TimeAxisIntervalType) {
    /**
     * Ticks are ten millionths of seconds.
     */
    TimeAxisIntervalType[TimeAxisIntervalType["Ticks"] = 0] = "Ticks";
    /**
     * Milliseconds.
     */
    TimeAxisIntervalType[TimeAxisIntervalType["Milliseconds"] = 1] = "Milliseconds";
    /**
     * Seconds.
     */
    TimeAxisIntervalType[TimeAxisIntervalType["Seconds"] = 2] = "Seconds";
    /**
     * Minutes.
     */
    TimeAxisIntervalType[TimeAxisIntervalType["Minutes"] = 3] = "Minutes";
    /**
     * Hours.
     */
    TimeAxisIntervalType[TimeAxisIntervalType["Hours"] = 4] = "Hours";
    /**
     * Days.
     */
    TimeAxisIntervalType[TimeAxisIntervalType["Days"] = 5] = "Days";
    /**
     * Weeks.
     */
    TimeAxisIntervalType[TimeAxisIntervalType["Weeks"] = 6] = "Weeks";
    /**
     * Months.
     */
    TimeAxisIntervalType[TimeAxisIntervalType["Months"] = 7] = "Months";
    /**
     * Years.
     */
    TimeAxisIntervalType[TimeAxisIntervalType["Years"] = 8] = "Years";
    return TimeAxisIntervalType;
})({});
/**
 * @hidden
 */
export let TimeAxisIntervalType_$type = /*@__PURE__*/ markEnum('TimeAxisIntervalType', 'Ticks,0|Milliseconds,1|Seconds,2|Minutes,3|Hours,4|Days,5|Weeks,6|Months,7|Years,8');
