// igniteui-react-charts/esm2015/lib/TooltipUpdateMessage.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RenderingMessage } from "./RenderingMessage";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let TooltipUpdateMessage = /*@__PURE__*/ (() => {
    class TooltipUpdateMessage extends RenderingMessage {
        constructor() {
            super(...arguments);
            this.h = null;
            this.f = null;
        }
        get i() {
            return this.h;
        }
        set i(a) {
            this.h = a;
        }
        get g() {
            return this.f;
        }
        set g(a) {
            this.f = a;
        }
        toString() {
            return "TooltipUpdateMessage[" + this.i.toString() + ", " + this.g.toString() + "]";
        }
    }
    TooltipUpdateMessage.$t = /*@__PURE__*/ markType(TooltipUpdateMessage, 'TooltipUpdateMessage', RenderingMessage.$);
    return TooltipUpdateMessage;
})();
