// igniteui-react-charts/esm2015/lib/ToolTipType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * An enum representing the available tooltip types to display in the chart.
 */
export var ToolTipType = /*@__PURE__*/ (function (ToolTipType) {
    /**
     * Display default tooltip for each series in the chart.
     */
    ToolTipType[ToolTipType["Default"] = 0] = "Default";
    /**
     * Display individual tooltips for all series in the chart.
     */
    ToolTipType[ToolTipType["Item"] = 1] = "Item";
    /**
     * Display combined tooltip for all series in the chart.
     */
    ToolTipType[ToolTipType["Category"] = 2] = "Category";
    /**
     * Display no tooltips in the chart.
     */
    ToolTipType[ToolTipType["None"] = 3] = "None";
    /**
     * Display the data tooltips for all series in the chart
     */
    ToolTipType[ToolTipType["Data"] = 4] = "Data";
    return ToolTipType;
})({});
/**
 * @hidden
 */
export let ToolTipType_$type = /*@__PURE__*/ markEnum('ToolTipType', 'Default,0|Item,1|Category,2|None,3|Data,4');
