// igniteui-react-charts/esm2015/lib/TooltipMessage_Spark.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RenderingMessage_Spark } from "./RenderingMessage_Spark";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let TooltipMessage_Spark = /*@__PURE__*/ (() => {
    class TooltipMessage_Spark extends RenderingMessage_Spark {
        constructor() {
            super(...arguments);
            this.c = 0;
            this.d = 0;
            this.g = null;
        }
        get xOffset() {
            return this.c;
        }
        set xOffset(a) {
            this.c = a;
        }
        get yOffset() {
            return this.d;
        }
        set yOffset(a) {
            this.d = a;
        }
        get context() {
            return this.g;
        }
        set context(a) {
            this.g = a;
        }
    }
    TooltipMessage_Spark.$t = /*@__PURE__*/ markType(TooltipMessage_Spark, 'TooltipMessage_Spark', RenderingMessage_Spark.$);
    return TooltipMessage_Spark;
})();
