// igniteui-react-charts/esm2015/lib/TreemapNodeVisualData.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { StringBuilder } from "igniteui-react-core";
/**
 * @hidden
 */
export let TreemapNodeVisualData = /*@__PURE__*/ (() => {
    class TreemapNodeVisualData extends Base {
        constructor() {
            super(...arguments);
            this.d = null;
            this.b = null;
            this.c = null;
            this.a = null;
        }
        f(a) {
            if (this.d != null) {
                this.d.scaleByViewport(a);
            }
            if (this.b != null) {
                this.b.scaleByViewport(a);
            }
            if (this.a != null) {
                this.a.scaleByViewport(a);
            }
            if (this.c != null) {
                this.c.scaleByViewport(a);
            }
        }
        e() {
            let a = new StringBuilder(0);
            let b = true;
            a.u("{");
            if (this.d != null) {
                if (b) {
                    b = false;
                }
                else {
                    a.l(", ");
                }
                a.l("nodeRect: ");
                a.l(this.d.serialize());
            }
            if (this.b != null) {
                if (b) {
                    b = false;
                }
                else {
                    a.l(", ");
                }
                a.l("headerRect: ");
                a.l(this.b.serialize());
            }
            if (this.a != null) {
                if (b) {
                    b = false;
                }
                else {
                    a.l(", ");
                }
                a.l("label: ");
                a.l(this.a.serialize());
            }
            if (this.c != null) {
                if (b) {
                    b = false;
                }
                else {
                    a.l(", ");
                }
                a.l("highlightRect: ");
                a.l(this.c.serialize());
            }
            a.u("");
            a.u("}");
            return a.toString();
        }
    }
    TreemapNodeVisualData.$t = /*@__PURE__*/ markType(TreemapNodeVisualData, 'TreemapNodeVisualData');
    return TreemapNodeVisualData;
})();
