// igniteui-react-charts/esm2015/lib/TreemapVisual.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
/**
 * @hidden
 */
export let TreemapVisual = /*@__PURE__*/ (() => {
    class TreemapVisual extends Base {
        constructor() {
            super();
            this.a = null;
            this.l = null;
            this.o = null;
            this.m = null;
            this.r = null;
            this.s = null;
            this.j = null;
            this.n = null;
            this.p = null;
            this.x = 0;
            this.c = null;
            this.i = 0;
            this.b = false;
            this.f = 0;
            this.e = 0;
            this.u = 0;
            this.w = 0;
            this.v = 0;
            this.k = 0;
            this.t = 0;
            this.h = 0;
            this.q = null;
            this.g = 0;
            this.d = 0;
            this.c = new List$1(Base.$, 0);
            this.h = 1;
        }
    }
    TreemapVisual.$t = /*@__PURE__*/ markType(TreemapVisual, 'TreemapVisual');
    return TreemapVisual;
})();
