// igniteui-react-charts/esm2015/lib/TrendResolutionParams.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let TrendResolutionParams = /*@__PURE__*/ (() => {
    class TrendResolutionParams extends Base {
        constructor() {
            super(...arguments);
            this.f = 0;
            this.g = 0;
            this.e = 0;
            this.i = null;
            this.j = null;
            this.h = null;
            this.d = 0;
            this.a = 0;
            this.c = 0;
            this.b = 0;
        }
    }
    TrendResolutionParams.$t = /*@__PURE__*/ markType(TrendResolutionParams, 'TrendResolutionParams');
    return TrendResolutionParams;
})();
