// igniteui-react-charts/esm2015/lib/UltimateOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { UltimateOscillatorIndicatorCalculationStrategy } from "./UltimateOscillatorIndicatorCalculationStrategy";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let UltimateOscillatorIndicator = /*@__PURE__*/ (() => {
    class UltimateOscillatorIndicator extends StrategyBasedIndicator {
        bj() {
            return new UltimateOscillatorIndicator();
        }
        get_abv() {
            return new UltimateOscillatorIndicatorCalculationStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return UltimateOscillatorIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
    }
    UltimateOscillatorIndicator.$t = /*@__PURE__*/ markType(UltimateOscillatorIndicator, 'UltimateOscillatorIndicator', StrategyBasedIndicator.$);
    return UltimateOscillatorIndicator;
})();
