// igniteui-react-charts/esm2015/lib/ViewportChangedMessage_Spark.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { InteractionMessage_Spark } from "./InteractionMessage_Spark";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ViewportChangedMessage_Spark = /*@__PURE__*/ (() => {
    class ViewportChangedMessage_Spark extends InteractionMessage_Spark {
        constructor() {
            super(...arguments);
            this.d = 0;
            this.c = 0;
            this.e = 0;
            this.b = 0;
        }
        get newTop() {
            return this.d;
        }
        set newTop(a) {
            this.d = a;
        }
        get newLeft() {
            return this.c;
        }
        set newLeft(a) {
            this.c = a;
        }
        get newWidth() {
            return this.e;
        }
        set newWidth(a) {
            this.e = a;
        }
        get newHeight() {
            return this.b;
        }
        set newHeight(a) {
            this.b = a;
        }
        toString() {
            return "ViewportChangedMessage[" + this.newTop.toString() + ", " + this.newLeft.toString() + ", " + this.newWidth.toString() + ", " + this.newHeight.toString() + "]";
        }
    }
    ViewportChangedMessage_Spark.$t = /*@__PURE__*/ markType(ViewportChangedMessage_Spark, 'ViewportChangedMessage_Spark', InteractionMessage_Spark.$);
    return ViewportChangedMessage_Spark;
})();
