// igniteui-react-charts/esm2015/lib/ViewportChangedMessage.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { InteractionMessage } from "./InteractionMessage";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ViewportChangedMessage = /*@__PURE__*/ (() => {
    class ViewportChangedMessage extends InteractionMessage {
        constructor() {
            super(...arguments);
            this.c = 0;
            this.b = 0;
        }
        get e() {
            return this.c;
        }
        set e(a) {
            this.c = a;
        }
        get d() {
            return this.b;
        }
        set d(a) {
            this.b = a;
        }
        toString() {
            return "ViewportChangedMessage[" + this.e.toString() + ", " + this.d.toString() + "]";
        }
    }
    ViewportChangedMessage.$t = /*@__PURE__*/ markType(ViewportChangedMessage, 'ViewportChangedMessage', InteractionMessage.$);
    return ViewportChangedMessage;
})();
