// igniteui-react-charts/esm2015/lib/WilliamsPercentRIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { WilliamsPercentRIndicatorStrategy } from "./WilliamsPercentRIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let WilliamsPercentRIndicator = /*@__PURE__*/ (() => {
    class WilliamsPercentRIndicator extends StrategyBasedIndicator {
        bj() {
            return new WilliamsPercentRIndicator();
        }
        get_abv() {
            return new WilliamsPercentRIndicatorStrategy();
        }
        get abv() {
            return this.get_abv();
        }
        get_ab8() {
            return WilliamsPercentRIndicator.$;
        }
        get ab8() {
            return this.get_ab8();
        }
        get acg() {
            return typeGetValue(this.c(WilliamsPercentRIndicator.ach));
        }
        set acg(a) {
            this.h(WilliamsPercentRIndicator.ach, a);
        }
        ab1() {
            return this.acg;
        }
        aca(a) {
            this.acg = a;
        }
    }
    WilliamsPercentRIndicator.$t = /*@__PURE__*/ markType(WilliamsPercentRIndicator, 'WilliamsPercentRIndicator', StrategyBasedIndicator.$);
    WilliamsPercentRIndicator.ach = /*@__PURE__*/ StrategyBasedIndicator.acd(14, WilliamsPercentRIndicator.$);
    return WilliamsPercentRIndicator;
})();
