// igniteui-react-charts/esm2015/lib/WinLossColumnMessage_Spark.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RenderingMessage_Spark } from "./RenderingMessage_Spark";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let WinLossColumnMessage_Spark = /*@__PURE__*/ (() => {
    class WinLossColumnMessage_Spark extends RenderingMessage_Spark {
        constructor() {
            super(...arguments);
            this.d = null;
            this.g = null;
            this.k = null;
            this.l = null;
            this.f = null;
            this.e = null;
            this.i = 0;
            this.h = 0;
            this.j = 0;
            this.c = 0;
        }
    }
    WinLossColumnMessage_Spark.$t = /*@__PURE__*/ markType(WinLossColumnMessage_Spark, 'WinLossColumnMessage_Spark', RenderingMessage_Spark.$);
    return WinLossColumnMessage_Spark;
})();
