// igniteui-react-charts/esm2015/lib/XamPieChart_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PieChartBase } from "./PieChartBase";
import { PieChartBaseView } from "./PieChartBaseView";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let XamPieChart = /*@__PURE__*/ (() => {
    class XamPieChart extends PieChartBase {
        a3() {
            return new XamPieChartView(this);
        }
        go(a) {
            super.go(a);
            this.jf = a;
        }
        constructor() {
            super();
            this.jf = null;
            this.ab = XamPieChart.$;
        }
    }
    XamPieChart.$t = /*@__PURE__*/ markType(XamPieChart, 'XamPieChart', PieChartBase.$);
    return XamPieChart;
})();
/**
 * @hidden
 */
export let XamPieChartView = /*@__PURE__*/ (() => {
    class XamPieChartView extends PieChartBaseView {
        constructor(a) {
            super(a);
            this.b6 = null;
            this.b6 = a;
        }
    }
    XamPieChartView.$t = /*@__PURE__*/ markType(XamPieChartView, 'XamPieChartView', PieChartBaseView.$);
    return XamPieChartView;
})();
