// igniteui-react-charts/esm5/lib/AbsoluteVolumeOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { AbsoluteVolumeOscillatorIndicatorStrategy } from "./AbsoluteVolumeOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var AbsoluteVolumeOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AbsoluteVolumeOscillatorIndicator, _super);
    function AbsoluteVolumeOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AbsoluteVolumeOscillatorIndicator.prototype.bj = function () {
        return new AbsoluteVolumeOscillatorIndicator();
    };
    AbsoluteVolumeOscillatorIndicator.prototype.get_abv = function () {
        return new AbsoluteVolumeOscillatorIndicatorStrategy();
    };
    Object.defineProperty(AbsoluteVolumeOscillatorIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    AbsoluteVolumeOscillatorIndicator.prototype.get_ab8 = function () {
        return AbsoluteVolumeOscillatorIndicator.$;
    };
    Object.defineProperty(AbsoluteVolumeOscillatorIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AbsoluteVolumeOscillatorIndicator.prototype, "shortPeriod", {
        get: function () {
            return typeGetValue(this.c(AbsoluteVolumeOscillatorIndicator.ach));
        },
        set: function (a) {
            this.h(AbsoluteVolumeOscillatorIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    AbsoluteVolumeOscillatorIndicator.prototype.ab3 = function () {
        return this.shortPeriod;
    };
    Object.defineProperty(AbsoluteVolumeOscillatorIndicator.prototype, "longPeriod", {
        get: function () {
            return typeGetValue(this.c(AbsoluteVolumeOscillatorIndicator.acg));
        },
        set: function (a) {
            this.h(AbsoluteVolumeOscillatorIndicator.acg, a);
        },
        enumerable: false,
        configurable: true
    });
    AbsoluteVolumeOscillatorIndicator.prototype.ab0 = function () {
        return this.longPeriod;
    };
    AbsoluteVolumeOscillatorIndicator.$t = markType(AbsoluteVolumeOscillatorIndicator, 'AbsoluteVolumeOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    AbsoluteVolumeOscillatorIndicator.ach = StrategyBasedIndicator.acf(10, AbsoluteVolumeOscillatorIndicator.$);
    AbsoluteVolumeOscillatorIndicator.acg = StrategyBasedIndicator.acb(30, AbsoluteVolumeOscillatorIndicator.$);
    return AbsoluteVolumeOscillatorIndicator;
}(StrategyBasedIndicator));
export { AbsoluteVolumeOscillatorIndicator };
