// igniteui-react-charts/esm5/lib/AverageTrueRangeIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AverageTrueRangeIndicatorStrategy } from "./AverageTrueRangeIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var AverageTrueRangeIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AverageTrueRangeIndicator, _super);
    function AverageTrueRangeIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AverageTrueRangeIndicator.prototype.bj = function () {
        return new AverageTrueRangeIndicator();
    };
    AverageTrueRangeIndicator.prototype.get_abv = function () {
        return new AverageTrueRangeIndicatorStrategy();
    };
    Object.defineProperty(AverageTrueRangeIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    AverageTrueRangeIndicator.prototype.get_ab8 = function () {
        return AverageTrueRangeIndicator.$;
    };
    Object.defineProperty(AverageTrueRangeIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AverageTrueRangeIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(AverageTrueRangeIndicator.ach));
        },
        set: function (a) {
            this.h(AverageTrueRangeIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    AverageTrueRangeIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    AverageTrueRangeIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    AverageTrueRangeIndicator.$t = markType(AverageTrueRangeIndicator, 'AverageTrueRangeIndicator', StrategyBasedIndicator.$);
    AverageTrueRangeIndicator.ach = StrategyBasedIndicator.acd(14, AverageTrueRangeIndicator.$);
    return AverageTrueRangeIndicator;
}(StrategyBasedIndicator));
export { AverageTrueRangeIndicator };
