// igniteui-react-charts/esm5/lib/AverageDirectionalIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AverageDirectionalIndexIndicatorStrategy } from "./AverageDirectionalIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var AverageDirectionalIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AverageDirectionalIndexIndicator, _super);
    function AverageDirectionalIndexIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AverageDirectionalIndexIndicator.prototype.bj = function () {
        return new AverageDirectionalIndexIndicator();
    };
    AverageDirectionalIndexIndicator.prototype.get_abv = function () {
        return new AverageDirectionalIndexIndicatorStrategy();
    };
    Object.defineProperty(AverageDirectionalIndexIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    AverageDirectionalIndexIndicator.prototype.get_ab8 = function () {
        return AverageDirectionalIndexIndicator.$;
    };
    Object.defineProperty(AverageDirectionalIndexIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AverageDirectionalIndexIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(AverageDirectionalIndexIndicator.ach));
        },
        set: function (a) {
            this.h(AverageDirectionalIndexIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    AverageDirectionalIndexIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    AverageDirectionalIndexIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    AverageDirectionalIndexIndicator.$t = markType(AverageDirectionalIndexIndicator, 'AverageDirectionalIndexIndicator', StrategyBasedIndicator.$);
    AverageDirectionalIndexIndicator.ach = StrategyBasedIndicator.acd(14, AverageDirectionalIndexIndicator.$);
    return AverageDirectionalIndexIndicator;
}(StrategyBasedIndicator));
export { AverageDirectionalIndexIndicator };
