// igniteui-react-charts/esm5/lib/BollingerBandWidthIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { BollingerBandWidthIndicatorStrategy } from "./BollingerBandWidthIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var BollingerBandWidthIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BollingerBandWidthIndicator, _super);
    function BollingerBandWidthIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BollingerBandWidthIndicator.prototype.bj = function () {
        return new BollingerBandWidthIndicator();
    };
    BollingerBandWidthIndicator.prototype.get_abv = function () {
        return new BollingerBandWidthIndicatorStrategy();
    };
    Object.defineProperty(BollingerBandWidthIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    BollingerBandWidthIndicator.prototype.get_ab8 = function () {
        return BollingerBandWidthIndicator.$;
    };
    Object.defineProperty(BollingerBandWidthIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BollingerBandWidthIndicator.prototype, "ach", {
        get: function () {
            return typeGetValue(this.c(BollingerBandWidthIndicator.ack));
        },
        set: function (a) {
            this.h(BollingerBandWidthIndicator.ack, a);
        },
        enumerable: false,
        configurable: true
    });
    BollingerBandWidthIndicator.prototype.ab1 = function () {
        return this.ach;
    };
    Object.defineProperty(BollingerBandWidthIndicator.prototype, "acg", {
        get: function () {
            return this.c(BollingerBandWidthIndicator.acj);
        },
        set: function (a) {
            this.h(BollingerBandWidthIndicator.acj, a);
        },
        enumerable: false,
        configurable: true
    });
    BollingerBandWidthIndicator.prototype.abz = function () {
        return this.acg;
    };
    BollingerBandWidthIndicator.$t = markType(BollingerBandWidthIndicator, 'BollingerBandWidthIndicator', StrategyBasedIndicator.$);
    BollingerBandWidthIndicator.ack = StrategyBasedIndicator.acd(20, BollingerBandWidthIndicator.$);
    BollingerBandWidthIndicator.acj = StrategyBasedIndicator.acc(2, BollingerBandWidthIndicator.$);
    return BollingerBandWidthIndicator;
}(StrategyBasedIndicator));
export { BollingerBandWidthIndicator };
