// igniteui-react-charts/esm5/lib/BubbleSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { ScatterBase } from "./ScatterBase";
import { IScaleLegendSeries_$type } from "igniteui-react-core";
import { ISupportsSizeScale_$type } from "./ISupportsSizeScale";
import { ISupportsFillScale_$type } from "./ISupportsFillScale";
import { Base, typeCast, Point_$type, fromEnum, markType, markDep, enumGetBox, runOn } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { UIElement } from "igniteui-react-core";
import { ActualRange } from "./ActualRange";
import { DependencyProperty } from "igniteui-react-core";
import { SizeScale } from "./SizeScale";
import { BrushScale } from "./BrushScale";
import { ScatterFrame } from "./ScatterFrame";
import { IChartLegend_$type } from "igniteui-react-core";
import { ValueBrushScale } from "./ValueBrushScale";
import { IFastItemsSource_$type } from "igniteui-react-core";
import { ScatterBaseView } from "./ScatterBaseView";
import { ScalerParams } from "./ScalerParams";
import { Dictionary$2 } from "igniteui-react-core";
import { OwnedPoint } from "./OwnedPoint";
import { BubbleMarkerManager } from "./BubbleMarkerManager";
import { IChartItemLegend_$type } from "igniteui-react-core";
import { IChartScaleLegend_$type } from "igniteui-react-core";
import { ContentControl } from "igniteui-react-core";
import { DataContext } from "igniteui-react-core";
import { DeviceUtils } from "igniteui-react-core";
import { BrushUtil } from "igniteui-react-core";
import { ChartSelection } from "./ChartSelection";
import { FastItemsSource } from "igniteui-react-core";
import { FastItemColumn } from "igniteui-react-core";
import { DataLegendSeriesValueInfo } from "igniteui-react-core";
import { DataLegendSeriesContext } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { isNaN_, isInfinity, logBase } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
import { BrushSelectionMode_$type } from "./BrushSelectionMode";
import { ScatterTrendLineManager } from "./ScatterTrendLineManager";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
var BubbleSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BubbleSeries, _super);
    function BubbleSeries() {
        var _this = _super.call(this) || this;
        _this.aeh = null;
        _this.aeg = null;
        _this.ac4 = null;
        _this.acv = null;
        _this.acu = null;
        _this.act = null;
        _this.acr = null;
        _this.ab = BubbleSeries.$;
        _this.y2 = new ScatterFrame();
        _this.y4 = new ScatterFrame();
        _this.y0 = new ScatterFrame();
        return _this;
    }
    BubbleSeries.prototype.bj = function () {
        return new BubbleSeries();
    };
    BubbleSeries.prototype.get_ei = function () {
        return true;
    };
    BubbleSeries.prototype.cs = function () {
        return new BubbleSeriesView(this);
    };
    BubbleSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.acq = a;
    };
    BubbleSeries.prototype.jl = function () {
        return 0.7;
    };
    BubbleSeries.prototype.w1 = function () {
        if (this.w0 == 0) {
            return 2;
        }
        return _super.prototype.w1.call(this);
    };
    BubbleSeries.prototype.isAttachedTo = function (a) {
        return this.a5 == a;
    };
    BubbleSeries.prototype.eh = function () {
        return true;
    };
    BubbleSeries.prototype.p2 = function (a) {
        _super.prototype.p2.call(this, a);
        if (this.acz != null && this.ac4 != null && this.ac4.count > 0) {
            for (var b = 0; b < this.ac4.count; b++) {
                a(this.ac4._inner[b]);
            }
        }
    };
    Object.defineProperty(BubbleSeries.prototype, "acy", {
        get: function () {
            return this.ac0;
        },
        enumerable: false,
        configurable: true
    });
    BubbleSeries.prototype.get_isColoredItemwise = function () {
        return this.acw != null && this.acw.count > 0;
    };
    BubbleSeries.prototype.aef = function (a) {
        var b = typeCast(CustomPaletteBrushScale.$, this.acp);
        var c = typeCast(ValueBrushScale.$, this.acp);
        var d = this.acp;
        if (b != null && this.dn != null) {
            return b.getBrush1(a, this.dn.count);
        }
        else if (c != null) {
            return c.getBrushByIndex(a, this.acw);
        }
        else if (d != null) {
            return d.getBrush(a);
        }
        return null;
    };
    Object.defineProperty(BubbleSeries.prototype, "legendReady", {
        get: function () {
            return this.ac0 != null && this.ac0.count != 0;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeries.prototype, "minScaleText", {
        get: function () {
            var a = this.aco(this.acs).b;
            return (Math.round(a * 1000) / 1000).toString();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeries.prototype, "maxScaleText", {
        get: function () {
            var a = this.aco(this.acs).a;
            return (Math.round(a * 1000) / 1000).toString();
        },
        enumerable: false,
        configurable: true
    });
    BubbleSeries.prototype.forScaleColors = function (a) {
        for (var b = 0; b < this.ac0.count; b++) {
            if (this.acp == null || !this.acp.isReady) {
                return false;
            }
            var c = NaN;
            if (this.acw != null) {
                c = (this.acw.item(b) - this.acw.minimum) / (this.acw.maximum - this.acw.minimum);
            }
            var d = (this.ac0.item(b) - this.ac0.minimum) / (this.ac0.maximum - this.ac0.minimum);
            if (isNaN_(c)) {
                c = 0;
            }
            if (isNaN_(d)) {
                d = c;
            }
            var e = this.yh;
            var f = this.aef(b);
            var g = f != null ? f : e;
            a(g, d);
        }
        return true;
    };
    BubbleSeries.prototype.kq = function (a, b) {
        var c = this.nd(b);
        if (c == BubbleSeries.$$p[5]) {
            return this.kr(a, b, this.adj);
        }
        if (c == BubbleSeries.$$p[11]) {
            return this.kr(a, b, this.adt);
        }
        if (c == ScatterBase.$$p[12]) {
            return this.kr(a, b, this.aat);
        }
        if (c == ScatterBase.$$p[16]) {
            return this.kr(a, b, this.aa1);
        }
        if (c == BubbleSeries.$$p[2]) {
            return this.kr(a, b, this.ade);
        }
        return _super.prototype.kq.call(this, a, b);
    };
    BubbleSeries.prototype.md = function (a) {
        var b = this.nd(a);
        if (b == BubbleSeries.$$p[5]) {
            return this.adj;
        }
        if (b == BubbleSeries.$$p[11]) {
            return this.adt;
        }
        if (b == ScatterBase.$$p[12]) {
            return this.aat;
        }
        if (b == ScatterBase.$$p[16]) {
            return this.aa1;
        }
        if (b == BubbleSeries.$$p[2]) {
            return this.ade;
        }
        return _super.prototype.md.call(this, a);
    };
    Object.defineProperty(BubbleSeries.prototype, "ac0", {
        get: function () {
            return this.acv;
        },
        set: function (a) {
            if (this.acv != a) {
                var b = this.ac0;
                this.acv = a;
                this.raisePropertyChanged("RadiusColumn", b, this.ac0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeries.prototype, "acz", {
        get: function () {
            return this.acu;
        },
        set: function (a) {
            if (this.acu != a) {
                var b = this.acz;
                this.acu = a;
                this.raisePropertyChanged("LabelColumn", b, this.acz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeries.prototype, "acw", {
        get: function () {
            return this.act;
        },
        set: function (a) {
            if (this.act != a) {
                var b = this.acw;
                this.act = a;
                this.raisePropertyChanged("FillColumn", b, this.acw);
            }
        },
        enumerable: false,
        configurable: true
    });
    BubbleSeries.prototype.aa4 = function (a, b, c, d, e) {
        if (c <= this.z9) {
            b.g = new Dictionary$2(Base.$, OwnedPoint.$, 1, c);
        }
        var f = this.z9;
        var g = this.dn;
        var h;
        var i;
        var j = this.getEffectiveViewport1(a);
        var k = new ScalerParams(0, d, e, this.yw.d, j);
        var l = new ScalerParams(0, d, e, this.yw.e, j);
        var m = false;
        var n = 0;
        if (this.db != null) {
            if (this.db.f4()) {
                var o = this.db.m4(d, k);
                m = o.c;
                n = o.d;
            }
        }
        var _loop_1 = function (p) {
            h = this_1.zh.item(p);
            i = this_1.zi.item(p);
            var q = { $type: Point_$type, x: this_1.yw.a.el(h, k), y: this_1.yw.b.el(i, l) };
            if (m) {
                if (q.x < e.left || q.x > e.right) {
                    if (q.x + n >= e.left && q.x + n <= e.right) {
                        q = { $type: Point_$type, x: q.x + n, y: i };
                    }
                }
            }
            if (!isInfinity(q.x) && !isInfinity(q.y)) {
                var r_1 = g.item(p);
                if (c <= f) {
                    if (!b.g.containsKey(r_1)) {
                        var s_1 = { $type: Point_$type, x: h, y: i };
                        b.g.addItem(r_1, ((function () {
                            var $ret = new OwnedPoint();
                            $ret.a = r_1;
                            $ret.b = s_1;
                            $ret.c = q;
                            return $ret;
                        })()));
                    }
                }
            }
        };
        var this_1 = this;
        for (var p = 0; p < c; p++) {
            _loop_1(p);
        }
    };
    BubbleSeries.prototype.abf = function (a, b) {
        var c = typeCast(BubbleSeriesView.$, b);
        var d = c.c0;
        d.m = this.ac0;
        _super.prototype.abf.call(this, a, b);
        var e = Math.min(this.zh != null ? this.zh.count : 0, this.zi != null ? this.zi.count : 0);
        if (e < 1) {
            return;
        }
        c.dm();
        c.dn();
    };
    BubbleSeries.prototype.ady = function () {
        if (this.db == null) {
            return;
        }
        if (this.a5 == null) {
            return;
        }
        if (this.a5.isItemwise) {
            var a = this.a5;
            this.a5.clearLegendItemsForSeries(this);
            this.adx();
            a.renderItemwiseContent(this);
        }
        if (this.a5.isScale) {
            var b = this.a5;
            b.restoreOriginalState();
            b.initializeLegend(this);
        }
    };
    BubbleSeries.ac5 = function (a, b, c, d, e) {
        if (e <= a || isNaN_(e) || isInfinity(e)) {
            return c;
        }
        if (e >= b) {
            return d;
        }
        var f = c + ((d - c) / (b - a)) * (e - a);
        return f;
    };
    BubbleSeries.ac6 = function (a, b, c, d, logBase_, e) {
        var f = logBase(e, logBase_);
        var g = logBase(a, logBase_);
        var h = logBase(b, logBase_);
        return BubbleSeries.ac5(g, h, c, d, f);
    };
    BubbleSeries.prototype.he = function (a, b, c) {
        var d = _super.prototype.he.call(this, a, b, c);
        if (this.ac0 == null || this.dn == null || this.ac0.count == 0 || this.dn.count != this.ac0.count) {
            d = false;
        }
        return d;
    };
    BubbleSeries.prototype.zr = function (a, b) {
        if (b != 3) {
            return true;
        }
        if (a == null) {
            return true;
        }
        if (this.aat == a || this.aa1 == a || this.adt == a) {
            return true;
        }
        return false;
    };
    BubbleSeries.prototype.adx = function () {
        var _this = this;
        var a = typeCast(IChartLegend_$type, this.a5);
        if (a == null || this.dn == null) {
            return;
        }
        this.ac4 = new List$1(UIElement.$, 0);
        var b = typeCast(CustomPaletteBrushScale.$, this.acp);
        var c = typeCast(ValueBrushScale.$, this.acp);
        var d = this.acp;
        var e = this.acn();
        var f = e.b;
        var g = e.a;
        var h = this.dn;
        var _loop_2 = function (i) {
            var j = null;
            if (c != null && this_2.acw != null) {
                var k = this_2.acw.item(i);
                j = c.getBrushByRange(k, f, g);
            }
            else if (b != null) {
                j = b.getBrush1(i, h.count);
            }
            else if (d != null) {
                j = d.getBrush(i);
            }
            var l = new ContentControl();
            var m = this_2.acz != null && this_2.acz.item(i) != null ? this_2.acz.item(i).toString() : "";
            l.content = ((function () {
                var $ret = new DataContext();
                $ret.series = _this;
                $ret.item = h.item(i);
                $ret.itemBrush = j;
                $ret.itemLabel = m;
                $ret.legendLabel = m;
                return $ret;
            })());
            var n = this_2.sx;
            if (n == null) {
                n = this_2.cv.bn(this_2.c0);
            }
            l.ah = n;
            this_2.ac4.add(l);
        };
        var this_2 = this;
        for (var i = 0; i < h.count; i++) {
            _loop_2(i);
        }
    };
    BubbleSeries.prototype.pb = function (a, b, c, d) {
        _super.prototype.pb.call(this, a, b, c, d);
        if (this.acs != null) {
            this.acs.aa();
        }
        if (this.acp != null) {
            this.acp.x();
        }
        this.ady();
    };
    BubbleSeries.prototype.qd = function () {
        _super.prototype.qd.call(this);
        this.ady();
    };
    BubbleSeries.prototype.qk = function () {
        _super.prototype.qk.call(this);
        this.ady();
    };
    BubbleSeries.prototype.q9 = function (a, b, c, d) {
        var e_1, _a;
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case ScatterBase.$$p[10]:
            case ScatterBase.$$p[14]:
            case ScatterBase.$$p[11]:
            case ScatterBase.$$p[15]:
            case BubbleSeries.$$p[9]:
            case BubbleSeries.$$p[10]:
                this.qf();
                break;
            case "FastItemsSource":
                if (this.acs != null) {
                    this.acs.aa();
                }
                if (this.acp != null) {
                    this.acp.x();
                }
                if (typeCast(IFastItemsSource_$type, c) != null) {
                    c.deregisterColumn(this.ac0);
                    c.deregisterColumn(this.acw);
                    c.deregisterColumn(this.acz);
                    this.ac0 = null;
                    this.acw = null;
                    this.acz = null;
                }
                if (typeCast(IFastItemsSource_$type, d) != null) {
                    this.ac0 = this.dk(this.adt);
                    if (!stringIsNullOrEmpty(this.ade)) {
                        this.acw = this.dk(this.ade);
                    }
                    this.acz = this.dm(this.adj);
                }
                this.rr(false);
                this.ady();
                break;
            case BubbleSeries.$$p[11]:
                if (this.acs != null) {
                    this.acs.aa();
                }
                if (this.dn != null) {
                    this.dn.deregisterColumn(this.ac0);
                    this.ac0 = this.dk(this.adt);
                    this.ady();
                }
                break;
            case "RadiusColumn":
                if (this.acs != null) {
                    this.acs.aa();
                }
                this.yx.c2.g$i();
                this.rr(this.gg && this.d7);
                this.qm();
                break;
            case BubbleSeries.$$p[12]:
                if (c != null) {
                    c.l.remove(this);
                    c.aa();
                }
                if (d != null) {
                    d.l.add(this);
                    d.aa();
                }
                this.rr(false);
                this.ql();
                break;
            case BubbleSeries.$$p[13]:
                this.rr(false);
                this.ql();
                break;
            case "DiscreteLegendItemTemplate":
                this.ady();
                break;
            case BubbleSeries.$$p[3]:
                if (c != null) {
                    c.o.remove(this);
                    c.x();
                }
                if (d != null) {
                    d.o.add(this);
                    d.x();
                }
                this.rr(false);
                this.ady();
                this.ql();
                break;
            case BubbleSeries.$$p[4]:
            case BubbleSeries.$$p[8]:
            case BubbleSeries.$$p[7]:
            case BubbleSeries.$$p[6]:
                this.rr(false);
                this.ady();
                this.ql();
                break;
            case "FillColumn":
                this.rr(this.gg && this.d7);
                this.ql();
                break;
            case BubbleSeries.$$p[2]:
                if (this.acp != null) {
                    this.acp.x();
                }
                if (this.dn != null) {
                    this.dn.deregisterColumn(this.acw);
                    this.acw = this.dk(this.ade);
                    this.ady();
                }
                break;
            case BubbleSeries.$$p[5]:
                if (this.dn != null) {
                    this.dn.deregisterColumn(this.acz);
                    this.acz = this.dm(this.adj);
                    this.ady();
                }
                break;
            case "ActualLegend":
                var e = typeCast(IChartLegend_$type, c);
                if (e != null && e.isItemwise) {
                    var f = typeCast(IChartItemLegend_$type, e);
                    f.clearLegendItemsForSeries(this);
                }
                if (e != null && e.isScale) {
                    var g = typeCast(IChartScaleLegend_$type, e);
                    var h = true;
                    var i = null;
                    if (this.db != null) {
                        try {
                            for (var _b = __values(fromEnum(this.db.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                                var j = _c.value;
                                if (j.legend == g) {
                                    i = j;
                                    h = false;
                                }
                            }
                        }
                        catch (e_1_1) {
                            e_1 = { error: e_1_1 };
                        }
                        finally {
                            try {
                                if (_c && !_c.done && (_a = _b.return))
                                    _a.call(_b);
                            }
                            finally {
                                if (e_1)
                                    throw e_1.error;
                            }
                        }
                    }
                    if (h) {
                        g.restoreOriginalState();
                    }
                    else {
                        g.initializeLegend(typeCast(IScaleLegendSeries_$type, i));
                    }
                }
                this.ady();
                break;
            case "SeriesViewer":
                this.ady();
                break;
        }
    };
    BubbleSeries.prototype.updateSizeScaleGlobals = function () {
        var a = new ActualRange();
        a.b = NaN;
        a.a = NaN;
        if (this.ac0 != null) {
            a.b = this.ac0.minimum;
            a.a = this.ac0.maximum;
        }
        if (this.acs != null) {
            this.acs.ab(a.b, a.a);
        }
    };
    BubbleSeries.prototype.updateFillScaleGlobals = function () {
        var a = new ActualRange();
        a.b = NaN;
        a.a = NaN;
        if (this.acw != null) {
            a.b = this.acw.minimum;
            a.a = this.acw.maximum;
        }
        if (this.acp != null) {
            this.acp.z(a.b, a.a);
        }
    };
    BubbleSeries.prototype.aco = function (a) {
        var b = new ActualRange();
        b.b = NaN;
        b.a = NaN;
        if (this.ac0 != null) {
            b.b = this.ac0.minimum;
            b.a = this.ac0.maximum;
        }
        if (a != null) {
            a.x();
            if (this.ac3) {
                b.b = a.globalMinimum;
                b.a = a.globalMaximum;
            }
        }
        return b;
    };
    BubbleSeries.prototype.acn = function () {
        var a = new ActualRange();
        a.b = NaN;
        a.a = NaN;
        if (this.acw != null) {
            a.b = this.acw.minimum;
            a.a = this.acw.maximum;
        }
        var b = typeCast(ValueBrushScale.$, this.acp);
        if (b != null) {
            b.s();
            if (this.ac1 && this.acp.o.count > 1) {
                a.b = b.q;
                a.a = b.p;
            }
        }
        return a;
    };
    BubbleSeries.prototype.ad0 = function (a, b, c, d) {
        var e = c.width;
        if (!this.cv.bx.isEmpty) {
            e = this.cv.bx.width;
        }
        else if (this.db != null && !this.db.wk.isEmpty) {
            e = this.db.wk.width;
        }
        var f = DeviceUtils.c(e);
        this.acr = this.acs || ((function () {
            var $ret = new SizeScale();
            $ret.minimumValue = f / 10;
            $ret.maximumValue = f / 5;
            return $ret;
        })());
        if (this.acr != null) {
            var g = this.acr.minimumValue;
            var h = this.acr.maximumValue;
            var i = this.acr.logarithmBase;
            var j = this.aco(this.acr);
            var k = j.b;
            var l = j.a;
            if (this.acr.isLogarithmic) {
                for (var m = 0; m < b.count; m++) {
                    b._inner[m] = BubbleSeries.ac6(k, l, g, h, i, b._inner[m]);
                }
            }
            else {
                for (var n = 0; n < b.count; n++) {
                    b._inner[n] = BubbleSeries.ac5(k, l, g, h, b._inner[n]);
                }
            }
        }
        var o = 1;
        o = DeviceUtils.d();
        var p = this.acs == null;
        if (d) {
            var q = c.width / e;
            for (var r = 0; r < b.count; r++) {
                if (p) {
                    b._inner[r] = b._inner[r] / o;
                }
                b._inner[r] = b._inner[r] * q;
            }
        }
        for (var s = 0; s < a.count; s++) {
            var t = a._inner[s];
            if (p && !d) {
                t.width = Math.max(0, b._inner[s] / o);
                t.height = Math.max(0, b._inner[s] / o);
            }
            else {
                t.width = Math.max(0, b._inner[s]);
                t.height = Math.max(0, b._inner[s]);
            }
        }
    };
    BubbleSeries.prototype.adz = function (a) {
        var b = typeCast(BubbleSeriesView.$, this.cu);
        this.acq.b7 = false;
        if (b != null) {
            b.b7 = false;
        }
        var c = typeCast(CustomPaletteBrushScale.$, this.acp);
        var d = typeCast(ValueBrushScale.$, this.acp);
        var e = this.acp;
        var f = this.acn();
        var g = f.b;
        var h = f.a;
        var i = this.acp == null || !this.acp.isReady || (d != null && this.ade == null);
        if (i) {
            this.acq.dl();
            if (b != null) {
                b.dl();
            }
            return;
        }
        this.acq.b7 = true;
        if (b != null) {
            b.b7 = true;
        }
        var j = a.count;
        var k = this.dn;
        for (var l = 0; l < j; l++) {
            var m = a._inner[l];
            var n = typeCast(DataContext.$, m.content);
            if (n != null) {
                var o = null;
                var p = k.indexOf(n.item);
                if (c != null) {
                    o = c.getBrush1(p, k.count);
                }
                else if (d != null && this.acw != null) {
                    var q = this.acw.item(p);
                    o = d.getBrushByRange(q, g, h);
                }
                else if (e != null) {
                    o = e.getBrush(p);
                }
                else {
                    if (this.yj == null) {
                        o = this.yh;
                    }
                    else {
                        o = this.yj;
                    }
                }
                var r = null;
                if (this.yk == null) {
                    r = this.yi;
                }
                else {
                    r = this.yk;
                }
                if (this.ac2 && o != null) {
                    r = o;
                }
                if (r != null && !isNaN_(this.ac8)) {
                    r = BrushUtil.m(r, this.ac8);
                }
                if (o != null && !isNaN_(this.ac7)) {
                    o = BrushUtil.m(o, this.ac7);
                }
                n.itemBrush = o;
                n.actualItemBrush = o;
                n.outline = r;
            }
        }
    };
    BubbleSeries.prototype.ak = function (a, b) {
        var c = new List$1(ChartSelection.$, 0);
        this.p4(a, c, b);
        return c.toArray();
    };
    BubbleSeries.prototype.p4 = function (a, b, c) {
        var d = 1;
        if (d == 1) {
            var e = this.acx();
            this.aa7(a, b, c, d, e);
        }
        else {
            _super.prototype.p4.call(this, a, b, c);
        }
    };
    BubbleSeries.prototype.acx = function () {
        var a = new FastItemColumn(typeCast(FastItemsSource.$, this.dn), "RadiusColumn", null, false);
        var b = this.cv.bx;
        var c = b.width;
        if (!this.cv.bx.isEmpty) {
            c = this.cv.bx.width;
        }
        else if (this.db != null && !this.db.wk.isEmpty) {
            c = this.db.wk.width;
        }
        var d = DeviceUtils.c(c);
        this.acr = this.acs || ((function () {
            var $ret = new SizeScale();
            $ret.minimumValue = d / 10;
            $ret.maximumValue = d / 5;
            return $ret;
        })());
        if (this.acr != null) {
            var e = this.acr.minimumValue;
            var f = this.acr.maximumValue;
            var g = this.acr.logarithmBase;
            var h = this.aco(this.acr);
            var i = h.b;
            var j = h.a;
            if (this.acr.isLogarithmic) {
                for (var k = 0; k < a.count; k++) {
                    a.item(k, BubbleSeries.ac6(i, j, e, f, g, a.item(k)));
                }
            }
            else {
                for (var l = 0; l < a.count; l++) {
                    a.item(l, BubbleSeries.ac5(i, j, e, f, a.item(l)));
                }
            }
        }
        var m = DeviceUtils.d();
        var n = this.acs == null;
        var o = this.cv == this.cu;
        if (o) {
            var p = b.width / c;
            for (var q = 0; q < a.count; q++) {
                if (n) {
                    a.item(q, a.item(q) / m);
                }
                a.item(q, a.item(q) * p);
            }
        }
        for (var r = 0; r < a.count; r++) {
            var s = void 0;
            if (n && !o) {
                a.item(r, Math.max(0, a.item(r) / m));
            }
            else {
                a.item(r, Math.max(0, a.item(r)));
            }
        }
        return a;
    };
    BubbleSeries.prototype.getDataLegendSeriesContextAt = function (a, b) {
        var c = { $type: Point_$type, x: a, y: b };
        var d = this.kn(c);
        var e = d == null ? NaN : this.ix(d, "XMemberPath");
        var f = d == null ? NaN : this.ix(d, "YMemberPath");
        var g = new DataLegendSeriesValueInfo();
        g.a = 13;
        g.l = e;
        g.q = "X";
        g.r = this.aat != null ? this.aat : "X";
        g.s = this.aap != null ? this.aap : "X:";
        g.t = this.aar != null ? this.aar : "";
        var h = new DataLegendSeriesValueInfo();
        h.l = f;
        h.a = 14;
        h.q = "Y";
        h.r = this.aa1 != null ? this.aa1 : "Y";
        h.s = this.aax != null ? this.aax : "Y:";
        h.t = this.aaz != null ? this.aaz : "";
        var i = new DataLegendSeriesContext();
        i.e = ((function () {
            var $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
            $ret.add(g);
            $ret.add(h);
            return $ret;
        })());
        i.h = this.mc();
        if (!stringIsNullOrEmpty(this.adt)) {
            var j = d == null ? NaN : this.ix(d, "RadiusMemberPath");
            var k = new DataLegendSeriesValueInfo();
            k.l = j;
            k.a = 11;
            k.q = "Radius";
            k.r = this.adt != null ? this.adt : "Radius";
            k.s = this.adp != null ? this.adp : "R:";
            k.t = this.adr != null ? this.adr : "";
            i.e.add(k);
        }
        else if (this.acr != null) {
            var l = new DataLegendSeriesValueInfo();
            l.l = this.acr.maximumValue;
            l.a = 11;
            l.q = "Radius";
            l.r = this.adt != null ? this.adt : "Radius";
            l.s = this.adp != null ? this.adp : "R:";
            l.t = this.adr != null ? this.adr : "";
            i.e.add(l);
        }
        if (!stringIsNullOrEmpty(this.ade)) {
            var m = d == null ? NaN : this.ix(d, "FillMemberPath");
            var n = new DataLegendSeriesValueInfo();
            n.l = m;
            n.a = 12;
            n.q = "Fill";
            n.r = this.ade != null ? this.ade : "Fill";
            n.s = this.ada != null ? this.ada : "F:";
            n.t = this.adc != null ? this.adc : "";
            i.e.add(n);
        }
        return i;
    };
    Object.defineProperty(BubbleSeries.prototype, "actualMarkerBrush", {
        get: function () {
            return this.yh;
        },
        enumerable: false,
        configurable: true
    });
    BubbleSeries.$t = markType(BubbleSeries, 'BubbleSeries', ScatterBase.$, [IScaleLegendSeries_$type, ISupportsSizeScale_$type, ISupportsFillScale_$type]);
    BubbleSeries.$$p = markDep(DependencyProperty, PropertyMetadata, BubbleSeries, 'raisePropertyChanged', ['FillMemberAsLegendLabel:ada:ad1', [2, null], 'FillMemberAsLegendUnit:adc:ad2', [2, null], 'FillMemberPath:ade:ad3', [2], 'FillScale:acp:ad4', [BrushScale.$], 'FillScaleUseGlobalValues:ac1:ad5', [0, false], 'LabelMemberPath:adj:ad6', [2], 'MarkerBrushBrightness:ac7:ad7', [1, NaN], 'MarkerOutlineBrightness:ac8:ad8', [1, NaN], 'MarkerOutlineUsesFillScale:ac2:ad9', [0, false], 'RadiusMemberAsLegendLabel:adp:aea', [2, null], 'RadiusMemberAsLegendUnit:adr:aeb', [2, null], 'RadiusMemberPath:adt:aec', [2, null], 'RadiusScale:acs:aed', [SizeScale.$], 'RadiusScaleUseGlobalValues:ac3:aee', [0, false]]);
    return BubbleSeries;
}(ScatterBase));
export { BubbleSeries };
/**
 * @hidden
 */
var CustomPaletteBrushScale = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CustomPaletteBrushScale, _super);
    function CustomPaletteBrushScale() {
        return _super.call(this) || this;
    }
    CustomPaletteBrushScale.prototype.getBrush1 = function (a, b) {
        if (this.brushes == null || this.brushes.count == 0) {
            return null;
        }
        if (this.brushSelectionMode == 0) {
            return _super.prototype.getBrush.call(this, a % this.brushes.count);
        }
        var c = BubbleSeries.ac5(0, b - 1, 0, this.brushes.count - 1, a);
        return this.ab(c);
    };
    CustomPaletteBrushScale.prototype.get_isReady = function () {
        return this.brushes != null && this.brushes.count > 0;
    };
    CustomPaletteBrushScale.$t = markType(CustomPaletteBrushScale, 'CustomPaletteBrushScale', BrushScale.$);
    CustomPaletteBrushScale.$$p = markDep(DependencyProperty, PropertyMetadata, CustomPaletteBrushScale, 'v', ['BrushSelectionMode', [BrushSelectionMode_$type, enumGetBox(BrushSelectionMode_$type, 0)]]);
    return CustomPaletteBrushScale;
}(BrushScale));
export { CustomPaletteBrushScale };
/**
 * @hidden
 */
var BubbleSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BubbleSeriesView, _super);
    function BubbleSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.dk = null;
        _this.dk = a;
        _this.c2 = new ScatterTrendLineManager();
        return _this;
    }
    BubbleSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.b4.xa = Defaults.e;
        }
    };
    BubbleSeriesView.prototype.cz = function () {
        var _this = this;
        return new BubbleMarkerManager(function (a) { return _this.c3.item(a); }, function (a) { return _this.c1.yw.c.item(a); }, runOn(this, this.de), runOn(this, this.cw), runOn(this, this.c4));
    };
    BubbleSeriesView.prototype.dm = function () {
        var a = this.c0;
        this.dk.ad0(a.n, a.o, this.bx, this == this.e.cu);
        this.as();
    };
    BubbleSeriesView.prototype.dn = function () {
        var a = this.c0;
        this.dk.adz(a.n);
    };
    BubbleSeriesView.prototype.dl = function () {
        var e_2, _a;
        var a = this.c0;
        try {
            for (var _b = __values(fromEnum(a.n)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var b = _c.value;
                var c = typeCast(DataContext.$, b.content);
                if (c != null) {
                    c.itemBrush = null;
                    c.outline = null;
                }
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
    };
    BubbleSeriesView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span";
        var b = this.e.ba.a$u(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.dk.title + "</span><br/><span>" + "(${item." + this.dk.aat + "}, ${item." + this.dk.aa1 + "})</span>";
        if (!stringIsNullOrEmpty(this.dk.adt)) {
            a += "<span>, Radius: ${item." + this.dk.adt + "}</span>";
        }
        a += "</div>";
        return a;
    };
    BubbleSeriesView.$t = markType(BubbleSeriesView, 'BubbleSeriesView', ScatterBaseView.$);
    return BubbleSeriesView;
}(ScatterBaseView));
export { BubbleSeriesView };
