// igniteui-react-charts/esm5/lib/CalloutPlacementPositions.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Indicates the allowed directions that will be considered to position the callouts near the target points
 */
export var CalloutPlacementPositions = /*@__PURE__*/ (function (CalloutPlacementPositions) {
    CalloutPlacementPositions[CalloutPlacementPositions["Top"] = 0] = "Top";
    CalloutPlacementPositions[CalloutPlacementPositions["Left"] = 1] = "Left";
    CalloutPlacementPositions[CalloutPlacementPositions["Right"] = 2] = "Right";
    CalloutPlacementPositions[CalloutPlacementPositions["Bottom"] = 3] = "Bottom";
    CalloutPlacementPositions[CalloutPlacementPositions["TopLeft"] = 4] = "TopLeft";
    CalloutPlacementPositions[CalloutPlacementPositions["TopRight"] = 5] = "TopRight";
    CalloutPlacementPositions[CalloutPlacementPositions["BottomLeft"] = 6] = "BottomLeft";
    CalloutPlacementPositions[CalloutPlacementPositions["BottomRight"] = 7] = "BottomRight";
    return CalloutPlacementPositions;
})({});
/**
 * @hidden
 */
export var CalloutPlacementPositions_$type = /*@__PURE__*/ markEnum('CalloutPlacementPositions', 'Top,0|Left,1|Right,2|Bottom,3|TopLeft,4|TopRight,5|BottomLeft,6|BottomRight,7');
