// igniteui-react-charts/esm5/lib/CategoryChartType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available types of data visualizations in the Category Chart
 */
export var CategoryChartType = /*@__PURE__*/ (function (CategoryChartType) {
    /**
     * Specifies category line series with markers at each data point
     */
    CategoryChartType[CategoryChartType["Line"] = 0] = "Line";
    /**
     * Specifies category area series
     */
    CategoryChartType[CategoryChartType["Area"] = 1] = "Area";
    /**
     * Specifies category column chart with vertical rectangles at each data point
     */
    CategoryChartType[CategoryChartType["Column"] = 2] = "Column";
    /**
     * Specifies category point chart with markers at each data point
     */
    CategoryChartType[CategoryChartType["Point"] = 3] = "Point";
    /**
     * Specifies category step line chart
     */
    CategoryChartType[CategoryChartType["StepLine"] = 4] = "StepLine";
    /**
     * Specifies category step area chart
     */
    CategoryChartType[CategoryChartType["StepArea"] = 5] = "StepArea";
    /**
     * Specifies category spline line series with markers at each data point
     */
    CategoryChartType[CategoryChartType["Spline"] = 6] = "Spline";
    /**
     * Specifies category spline area series
     */
    CategoryChartType[CategoryChartType["SplineArea"] = 7] = "SplineArea";
    /**
     * Specifies category waterfall chart
     */
    CategoryChartType[CategoryChartType["Waterfall"] = 8] = "Waterfall";
    /**
     * Specifies automatic selection of chart type based on suggestions from Data Adapter
     */
    CategoryChartType[CategoryChartType["Auto"] = 9] = "Auto";
    return CategoryChartType;
})({});
/**
 * @hidden
 */
export var CategoryChartType_$type = /*@__PURE__*/ markEnum('CategoryChartType', 'Line,0|Area,1|Column,2|Point,3|StepLine,4|StepArea,5|Spline,6|SplineArea,7|Waterfall,8|Auto,9');
