// igniteui-react-charts/esm5/lib/CategoryYAxis_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { IZoomableCategoryAxis_$type } from "./IZoomableCategoryAxis";
import { DependencyProperty } from "igniteui-react-core";
import { Rect } from "igniteui-react-core";
import { Base, typeCast, Array_$type, Number_$type, markType, markDep } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { NumericScaler } from "./NumericScaler";
import { VerticalAxisLabelPanel } from "./VerticalAxisLabelPanel";
import { AxisRangeChangedEventArgs } from "./AxisRangeChangedEventArgs";
import { MathUtil } from "igniteui-react-core";
import { LinearCategorySnapper } from "./LinearCategorySnapper";
import { NumericXAxis } from "./NumericXAxis";
import { LabelPosition } from "./LabelPosition";
import { PropertyMetadata } from "igniteui-react-core";
import { truncate, isNaN_, isInfinity } from "igniteui-react-core";
import { CategoryAxisBaseView } from "./CategoryAxisBaseView";
import { AxisDefaults } from "./AxisDefaults";
/**
 * @hidden
 */
var CategoryYAxis = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryYAxis, _super);
    function CategoryYAxis() {
        var _this = _super.call(this) || this;
        _this.mo = 0;
        _this.mp = 0;
        _this.mw = 1;
        _this.mv = 1;
        _this.k2 = new List$1(Number_$type, 0);
        _this.ab = CategoryYAxis.$;
        return _this;
    }
    CategoryYAxis.prototype.a4 = function () {
        return new CategoryYAxisView(this);
    };
    CategoryYAxis.prototype.hv = function (a) {
        _super.prototype.hv.call(this, a);
        this.mm = a;
    };
    CategoryYAxis.prototype.get_cp = function () {
        return true;
    };
    CategoryYAxis.prototype.ek = function () {
        return this.kb().height;
    };
    CategoryYAxis.prototype.mj = function (a, b, c) {
        return this.mk(a, b, c, true);
    };
    CategoryYAxis.prototype.ml = function (a) {
        if (typeCast(List$1.$.specialize(Array_$type), a) !== null) {
            return a.asArray();
        }
        return null;
    };
    CategoryYAxis.prototype.hp = function (a, b, c, d, e) {
        var f;
        var g = e.e;
        var h = e.d;
        var i = e.c;
        var j = e.a;
        var k = !i.isEmpty;
        var l = this.categoryMode == 0 ? this.lj - 1 : this.lj;
        if (l < 0) {
            l = 0;
        }
        var m = i.top;
        var n = i.height;
        var o = g.top;
        var p = g.height;
        var q = h.top;
        var r = h.height;
        var s = 0;
        var t = 1;
        var u = this.ml(a);
        var v = false;
        if (u != null) {
            v = true;
        }
        var w;
        for (var x = 0; x < b.count; x++) {
            w = b.item(x);
            for (var y = c; y < d; y++) {
                if (v) {
                    f = u[y];
                }
                else {
                    f = a.item(y);
                }
                if (k) {
                    var z = l >= 1 ? (f[w]) / (l) : l == 0 ? 0.5 : NaN;
                    if (!j) {
                        z = 1 - z;
                    }
                    z = m + n * (z - s) / t;
                    var aa = (z - e.d.top) / e.d.height;
                    aa = e.d.top + e.d.height * (aa - e.e.top) / e.e.height;
                    f[w] = aa;
                }
                else {
                    var ab = l >= 1 ? (f[w]) / (l) : l == 0 ? 0.5 : NaN;
                    if (!j) {
                        ab = 1 - ab;
                    }
                    ab = q + r * (ab - o) / p;
                    f[w] = ab;
                }
            }
        }
    };
    Object.defineProperty(CategoryYAxis.prototype, "mt", {
        get: function () {
            return this.c(CategoryYAxis.nc);
        },
        set: function (a) {
            this.h(CategoryYAxis.nc, a);
        },
        enumerable: false,
        configurable: true
    });
    CategoryYAxis.prototype.mn = function () {
        return !isNaN_(this.mt);
    };
    Object.defineProperty(CategoryYAxis.prototype, "mq", {
        get: function () {
            return this.mo;
        },
        set: function (a) {
            if (this.mq != a) {
                var b = this.mo;
                this.mo = a;
                this.h5("ActualInterval", b, this.mq);
            }
        },
        enumerable: false,
        configurable: true
    });
    CategoryYAxis.prototype.ex = function () {
        return this.mq;
    };
    Object.defineProperty(CategoryYAxis.prototype, "mu", {
        get: function () {
            return this.c(CategoryYAxis.nd);
        },
        set: function (a) {
            this.h(CategoryYAxis.nd, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryYAxis.prototype, "mr", {
        get: function () {
            return this.mp;
        },
        set: function (a) {
            if (this.mr != a) {
                var b = this.mp;
                this.mp = a;
                this.h5("ActualMinorInterval", b, this.mr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryYAxis.prototype, "my", {
        get: function () {
            return this.mw;
        },
        set: function (a) {
            this.mw = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryYAxis.prototype, "mx", {
        get: function () {
            return this.mv;
        },
        set: function (a) {
            this.mv = a;
        },
        enumerable: false,
        configurable: true
    });
    CategoryYAxis.prototype.el = function (a, b) {
        if (!b.c.isEmpty) {
            var c = this.ms(a, NumericScaler.ac, b.c);
            c = (c - b.d.top) / b.d.height;
            c = b.d.top + b.d.height * (c - b.e.top) / b.e.height;
            return c;
        }
        return this.ms(a, b.e, b.d);
    };
    CategoryYAxis.prototype.ms = function (a, b, c) {
        var d = this.categoryMode == 0 ? this.lj - 1 : this.lj;
        if (d < 0) {
            d = 0;
        }
        var e = d >= 1 ? (a) / (d) : d == 0 ? 0.5 : NaN;
        if (!this.ci) {
            e = 1 - e;
        }
        return c.top + c.height * (e - b.top) / b.height;
    };
    CategoryYAxis.prototype.eo = function (a, b) {
        if (!b.c.isEmpty) {
            var c = b.e.top + b.e.height * (a - b.d.top) / b.d.height;
            c = (c * b.d.height) + b.d.top;
            c = this.getUnscaledValue3(c, NumericScaler.ac, b.c, this.categoryMode);
            return c;
        }
        return this.getUnscaledValue3(a, b.e, b.d, this.categoryMode);
    };
    CategoryYAxis.prototype.ep = function (a, b, c) {
        if (!b.c.isEmpty) {
            var d = b.e.top + b.e.height * (a - b.d.top) / b.d.height;
            d = (d * b.d.height) + b.d.top;
            d = this.getUnscaledValue3(d, NumericScaler.ac, b.c, c);
            return d;
        }
        return this.getUnscaledValue3(a, b.e, b.d, c);
    };
    CategoryYAxis.prototype.getUnscaledValue3 = function (a, b, c, d) {
        var e = b.top + (a - c.top) * b.height / c.height;
        if (!this.ci) {
            e = 1 - e;
        }
        var f = d == 0 ? this.lj - 1 : this.lj;
        if (f < 0) {
            f = 0;
        }
        return e * f;
    };
    CategoryYAxis.prototype.m9 = function (a, b) {
        if (isNaN_(a) || isNaN_(b) || this.kv == null || this.kv.count == 0) {
            return;
        }
        var c = a / this.kv.count;
        var d = b / this.kv.count;
        this.bi.wl = new Rect(0, this.bi.wl.x, 1 - d, this.bi.wl.width, d - c);
    };
    CategoryYAxis.prototype.av = function () {
        return new VerticalAxisLabelPanel();
    };
    CategoryYAxis.prototype.d6 = function () {
        return this.mt;
    };
    CategoryYAxis.prototype.getCategorySize = function (a, b, c) {
        var d = this.kf(b, c);
        return d.height / (this.lj * a.height);
    };
    CategoryYAxis.prototype.k8 = function (a, b, c) {
        var d = this.lf(a, b, c);
        var e = 0;
        if (!isNaN_(this.le)) {
            e = Math.min(this.le, 1);
        }
        var f = 1 - 0.5 * d;
        var g = this.getCategorySize(a, b, c) * f / (this.ll - (this.ll - 1) * e);
        return g;
    };
    CategoryYAxis.prototype.getGroupCenter = function (a, b, c, d) {
        var e = 0.5;
        if (this.ll > 1) {
            var f = this.lf(b, c, d);
            var g = 0;
            if (!isNaN_(this.le)) {
                g = Math.min(this.le, 1);
            }
            var h = 1 - 0.5 * f;
            var i = h / (this.ll - (this.ll - 1) * g);
            var j = (h - i) / (this.ll - 1);
            e = 0.25 * f + 0.5 * i + a * j;
        }
        return this.getCategorySize(b, c, d) * e;
    };
    CategoryYAxis.prototype.m8 = function (a) {
        var b = this.bi != null ? this.bi.v4 : Rect.empty;
        b = b.copy();
        var c = this.kh;
        var d = new Rect(0, 0, 0, 1, 1);
        var e = this.ke();
        var f = new ScalerParams(0, d, d, this.ch, e);
        var g = !b.isEmpty && !c.isEmpty && this.kv != null ? this.kv.indexOf(a) : -1;
        var h = g > -1 ? this.el(g, f) : NaN;
        if (!isNaN_(h) && this.bi.fg) {
            if (!isNaN_(h)) {
                if (h < b.top + 0.1 * b.height) {
                    h = h + 0.4 * b.height;
                    b.y = h - 0.5 * b.height;
                }
                if (h > b.bottom - 0.1 * b.height) {
                    h = h - 0.4 * b.height;
                    b.y = h - 0.5 * b.height;
                }
            }
            this.bi.ry(b, false, true);
        }
    };
    CategoryYAxis.prototype.c3 = function () {
        if (this.cq && this.c4) {
            this.bc = this.bb();
        }
        if (this.kv == null) {
            return false;
        }
        var a = false;
        var b = this.kv.count;
        if (b != this.mx) {
            var c = new AxisRangeChangedEventArgs(0, 1, 1, this.mx, b);
            this.mx = b;
            this.h6(c);
            a = true;
        }
        this.nb();
        this.na();
        return a;
    };
    CategoryYAxis.prototype.getWindowZoomFromCategories = function (a) {
        var b = MathUtil.a(a) && a > 0;
        var c = this.kv.count;
        if (c > 0 && b) {
            var d = this.bi.wk.height;
            var e = isNaN_(d) || isInfinity(d);
            if (e) {
                return NaN;
            }
            var f = new Rect(0, 0, 0, 1, 1);
            var g = this.bi.wk;
            var h = this.getCategorySize(f, g, g);
            var i = isNaN_(h) || isInfinity(h);
            if (i) {
                return NaN;
            }
            var j = h * a / d;
            if (j >= 0 && j <= 1) {
                return j;
            }
        }
        return NaN;
    };
    CategoryYAxis.prototype.getWindowZoomFromItemSpan = function (a) {
        var b = MathUtil.a(a) && a > 0;
        var c = this.kv.count;
        if (c > 0 && b) {
            var d = this.bi.wk.height;
            var e = isNaN_(d) || isInfinity(d);
            if (e) {
                return NaN;
            }
            var f = new Rect(0, 0, 0, 1, 1);
            var g = this.bi.wk;
            var h = this.k8(f, g, g);
            var i = a / h;
            var j = Math.min(1, 1 / i);
            if (j > 0 && j <= 1) {
                return j;
            }
        }
        return NaN;
    };
    CategoryYAxis.prototype.nb = function () {
        if (this.kv == null || this.bi == null) {
            return;
        }
        var a = !isNaN_(this.bi.is) && !isNaN_(this.bi.iu);
        if (a) {
            return;
        }
        if (!isNaN_(this.zoomMaximumItemSpan) && !isInfinity(this.zoomMaximumItemSpan)) {
            var b = this.getWindowZoomFromItemSpan(this.zoomMaximumItemSpan);
            if (!isNaN_(b) && !isInfinity(b)) {
                this.bi.hl = b;
            }
        }
        else if (!isNaN_(this.zoomMaximumCategoryRange) && !isInfinity(this.zoomMaximumCategoryRange)) {
            var c = this.getWindowZoomFromCategories(this.zoomMaximumCategoryRange);
            if (!isNaN_(c) && !isInfinity(c)) {
                this.bi.hl = c;
            }
        }
    };
    CategoryYAxis.prototype.na = function () {
        if (this.kv == null || this.bi == null) {
            return;
        }
        var a = NaN;
        var b = NaN;
        var c = this.bi.hs();
        var d = this.bi.ht();
        var e = false;
        if (MathUtil.a(this.zoomToItemSpan)) {
            e = true;
            d = this.getWindowZoomFromItemSpan(this.zoomToItemSpan);
            d = Math.max(d, this.bi.hl);
        }
        else if (MathUtil.a(this.zoomToCategoryRange)) {
            e = true;
            d = this.getWindowZoomFromCategories(this.zoomToCategoryRange);
            d = Math.max(d, this.bi.hl);
        }
        if (MathUtil.a(this.zoomToCategoryStart) && this.zoomToCategoryStart >= 0 && this.zoomToCategoryStart < this.lk) {
            e = true;
            b = this.getWindowZoomFromCategories(this.zoomToCategoryStart);
            var f = d + a - 1;
            var g = d - f;
            if (f > 0 && g > this.bi.hl) {
                d = g;
                d = Math.max(d, this.bi.hl);
            }
        }
        if (e) {
            this.bi.rw(a, b, c, d);
        }
    };
    CategoryYAxis.prototype.hz = function (a, b, c, d) {
        _super.prototype.hz.call(this, a, b, c, d);
        switch (b) {
            case CategoryYAxis.$$p[0]:
            case CategoryYAxis.$$p[1]:
                this.nb();
                break;
            case CategoryYAxis.$$p[2]:
            case CategoryYAxis.$$p[3]:
            case CategoryYAxis.$$p[4]:
                this.na();
                break;
        }
    };
    CategoryYAxis.prototype.ky = function (a) {
        if (a == null) {
            return false;
        }
        var b = this.be();
        if (b == null) {
            return false;
        }
        return b.synchronizeVertically;
    };
    CategoryYAxis.prototype.ic = function (a) {
        _super.prototype.ic.call(this, a);
        var b = this.bi != null ? this.bi.v4 : Rect.empty;
        var c = this.kh;
        var d = this.kc();
        var e = this.kb();
        var f = new ScalerParams(0, b, c, this.ch, d);
        var g = this.a5.a4();
        var h = this.a5.a7();
        var i = this.a5.a5();
        var j = this.a5.a6();
        var k = this.a5.e();
        var l = this.a5.f();
        var m = this.a5.g();
        this.ip();
        this.hk(g);
        this.hk(h);
        this.hk(i);
        this.hk(j);
        this.dd.clear();
        this.de.clear();
        this.k2.clear();
        this.a5.az(this, b, c);
        if (b.isEmpty || c.isEmpty) {
            this.br.count = 0;
            this.bq.count = 0;
            this.bp.count = 0;
        }
        if (this.br.count == 0 && this.bq.count == 0 && this.a5.l()) {
            this.a5.ad();
        }
        if (this.az != null) {
            this.az.registerAxis(this);
        }
        if (this.itemsSource == null || this.kv == null || this.kv.count == 0) {
            this.br.count = 0;
            this.bq.count = 0;
            this.bp.count = 0;
            this.a5.ad();
            return;
        }
        if (!b.isEmpty && !c.isEmpty) {
            var n = this.eo(e.top, f);
            var o = this.eo(e.bottom, f);
            if (!this.ch) {
                n = Math.ceil(n);
                o = Math.floor(o);
            }
            else {
                n = Math.floor(n);
                o = Math.ceil(o);
            }
            var p = c.left;
            var q = 0;
            if (this.aj != null) {
                var r = typeCast(NumericXAxis.$, this.aj);
                if (r != null) {
                    var s = this.ii(p, q, b, c, d, c.left, c.right);
                    p = s.p0;
                    q = s.p1;
                }
            }
            this.iq(g, p, c, k, true);
            this.a5.au(q);
            var t = Math.min(n, o);
            var u = Math.max(n, o);
            var v = new LinearCategorySnapper(1, t, u, e.height, this.mt, this.categoryMode, this.e0());
            var w = v.g;
            w = this.d3(w, this.mn(), b, c, d);
            if (v.g != w) {
                v = new LinearCategorySnapper(1, t, u, e.height, w, this.categoryMode, this.e0());
            }
            var x = Math.floor((t - 0) / v.g);
            var y = Math.ceil((u - 0) / v.g);
            if (!isNaN_(x) && !isNaN_(y)) {
                var z = truncate(x);
                var aa = truncate(y);
                var ab = this.el(0 + z * v.g, f);
                var ac = Math.abs(this.el(v.g, f) - this.el(0, f));
                this.aw.bw = this.el(v.g, f);
                for (var ad = z; ad <= aa; ++ad) {
                    var ae = this.el(0 + (ad + 1) * v.g, f);
                    if (ab <= e.bottom) {
                        if (ad % 2 == 0) {
                            this.hu(h, ab, ae, c);
                        }
                        this.ht(i, ab, c, l, false);
                        this.k2.add(ab);
                    }
                    if (this.categoryMode != 0 && this.ll != 0 && this.cy) {
                        if (this.mu > 0 && this.mu < v.g) {
                            var af = ac * this.mu;
                            if (!this.ch) {
                                var ag = ab;
                                while (ag - af > ae) {
                                    ag -= af;
                                    if (Math.abs(ag - ae) < 0.0001) {
                                        break;
                                    }
                                    this.ht(j, ag, c, m, false);
                                }
                            }
                            else {
                                var ah = ab;
                                while (ah + af < ae) {
                                    ah += af;
                                    if (Math.abs(ah - ae) < 0.0001) {
                                        break;
                                    }
                                    this.ht(j, ah, c, m, false);
                                }
                            }
                            this.mr = this.mu;
                        }
                        else {
                            for (var ai = 0; ai < truncate(v.g); ai++) {
                                for (var aj = 0; aj < this.ll; aj++) {
                                    var ak = this.getGroupCenter(aj, b, c, d);
                                    if (!this.ch) {
                                        ak = -ak;
                                    }
                                    var al = this.el(ai + ad * v.g, f) + ak;
                                    this.ht(j, al, c, m, false);
                                }
                            }
                            this.mr = 0.5;
                        }
                    }
                    var am = ab;
                    if (this.categoryMode != 0) {
                        var an = this.el(ad * v.g + 1, f);
                        am = (ab + an) / 2;
                    }
                    if (am <= e.bottom && am >= e.top) {
                        var ao = 0;
                        if (v.g >= 1) {
                            ao = ad * truncate(Math.floor(v.g));
                        }
                        else {
                            if ((ad * v.g) * 2 % 2 == 0) {
                                ao = truncate(Math.floor(ad * v.g));
                            }
                            else {
                                ao = -1;
                            }
                        }
                        if (this.kv != null && ao < this.kv.count && ao >= 0) {
                            var ap = this.kv.item(ao);
                            var aq = this.f1(ap);
                            if (!isNaN_(am) && !isInfinity(am) && aq != null) {
                                if ((typeof aq === 'string') && Base.equalsStatic(aq, "")) {
                                }
                                else {
                                    this.dd.add1(aq);
                                    this.de.add(new LabelPosition(am));
                                }
                            }
                        }
                    }
                    ab = ae;
                }
            }
            this.mq = v.g;
            if ((this.az == null || this.az.visibility == 0) && this.aj != null) {
                if (this.az != null && (this.az.location == 6 || this.az.location == 7)) {
                    this.bi.n3();
                }
            }
            this.a5.a0(this.dd, this.de);
            this.l9();
        }
    };
    CategoryYAxis.prototype.get_a2 = function () {
        return 1;
    };
    Object.defineProperty(CategoryYAxis.prototype, "a2", {
        get: function () {
            return this.get_a2();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryYAxis.prototype, "isVertical", {
        get: function () {
            return this.cp;
        },
        enumerable: false,
        configurable: true
    });
    CategoryYAxis.prototype.j$d = function (a, b) {
        return this.eo.apply(this, arguments);
    };
    CategoryYAxis.prototype.j$c = function (a, b) {
        return this.el.apply(this, arguments);
    };
    CategoryYAxis.prototype.j$e = function (a, b, c, d, e) {
        this.hp.apply(this, arguments);
    };
    CategoryYAxis.prototype.j$f = function (a, b, c, d) {
        this.hq.apply(this, arguments);
    };
    CategoryYAxis.prototype.j$g = function (a, b, c, d) {
        this.hr.apply(this, arguments);
    };
    Object.defineProperty(CategoryYAxis.prototype, "j$b", {
        get: function () {
            return this.ch;
        },
        enumerable: false,
        configurable: true
    });
    CategoryYAxis.prototype.j$a = function () {
        return this.ag.apply(this, arguments);
    };
    CategoryYAxis.$t = markType(CategoryYAxis, 'CategoryYAxis', CategoryAxisBase.$, [IZoomableCategoryAxis_$type]);
    CategoryYAxis.nc = DependencyProperty.i("Interval", Number_$type, CategoryYAxis.$, new PropertyMetadata(2, NaN, function (a, b) {
        a.h5("Interval", b.oldValue, b.newValue);
        a.ib(false);
    }));
    CategoryYAxis.nd = DependencyProperty.i("MinorInterval", Number_$type, CategoryYAxis.$, new PropertyMetadata(2, NaN, function (a, b) {
        (typeCast(CategoryYAxis.$, a)).h5("MinorInterval", b.oldValue, b.newValue);
        (typeCast(CategoryYAxis.$, a)).ib(false);
    }));
    CategoryYAxis.$$p = markDep(DependencyProperty, PropertyMetadata, CategoryYAxis, 'h5', ['ZoomMaximumCategoryRange::ne', [1, NaN], 'ZoomMaximumItemSpan::nf', [1, NaN], 'ZoomToCategoryRange::ng', [1, NaN], 'ZoomToCategoryStart::nh', [1, NaN], 'ZoomToItemSpan::ni', [1, NaN]]);
    return CategoryYAxis;
}(CategoryAxisBase));
export { CategoryYAxis };
/**
 * @hidden
 */
var CategoryYAxisView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryYAxisView, _super);
    function CategoryYAxisView(a) {
        var _this = _super.call(this, a) || this;
        _this.bm = null;
        _this.bm = a;
        return _this;
    }
    CategoryYAxisView.prototype.al = function () {
        _super.prototype.al.call(this);
        this.a.fj = AxisDefaults.f;
    };
    CategoryYAxisView.$t = markType(CategoryYAxisView, 'CategoryYAxisView', CategoryAxisBaseView.$);
    return CategoryYAxisView;
}(CategoryAxisBaseView));
export { CategoryYAxisView };
