// igniteui-react-charts/esm5/lib/CategoryBucketCalculator_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { Base, typeCast, markType, Array_$type, Number_$type, Point_$type, enumGetBox, runOn, delegateCombine, fromEnum, markDep } from "igniteui-react-core";
import { IBucketizer_$type } from "./IBucketizer";
import { ArgumentNullException } from "igniteui-react-core";
import { Rect } from "igniteui-react-core";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { ISortingAxis_$type } from "./ISortingAxis";
import { ScalerParams } from "./ScalerParams";
import { Series } from "./Series";
import { truncate, intDivide, isNaN_, isInfinity } from "igniteui-react-core";
import { FramePreparer } from "./FramePreparer";
import { IIsCategoryBased_$type } from "./IIsCategoryBased";
import { List$1 } from "igniteui-react-core";
import { ISupportsMarkers_$type } from "./ISupportsMarkers";
import { IProvidesViewport_$type } from "./IProvidesViewport";
import { ISupportsErrorBars_$type } from "./ISupportsErrorBars";
import { CategoryFrame } from "./CategoryFrame";
import { PreparationParams } from "./PreparationParams";
import { IDateTimeAxis_$type } from "./IDateTimeAxis";
import { NumericAxisBase } from "./NumericAxisBase";
import { IHasCategoryTrendline_$type } from "./IHasCategoryTrendline";
import { IHasSingleValueCategory_$type } from "./IHasSingleValueCategory";
import { DefaultCategoryTrendlineHost } from "./DefaultCategoryTrendlineHost";
import { DefaultSingleValueProvider } from "./DefaultSingleValueProvider";
import { SingleValuesHolder } from "./SingleValuesHolder";
import { MarkerSeries } from "./MarkerSeries";
import { IHasCategoryModePreference_$type } from "./IHasCategoryModePreference";
import { ICategorySeriesInteractionManager_$type } from "./ICategorySeriesInteractionManager";
import { DVContainer } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { CategoryCollisionMode_$type } from "./CategoryCollisionMode";
import { CategoryTransitionInMode_$type } from "./CategoryTransitionInMode";
import { HighlightingInfo } from "igniteui-react-core";
import { CategoryTransitionSourceFramePreparer } from "./CategoryTransitionSourceFramePreparer";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { Marker } from "./Marker";
import { MarkerSeriesView } from "./MarkerSeriesView";
import { CategorySeriesRenderManager } from "./CategorySeriesRenderManager";
import { Pool$1 } from "igniteui-react-core";
import { SeriesLayerPropertyOverlay } from "./SeriesLayerPropertyOverlay";
import { DoubleCollection } from "igniteui-react-core";
import { Clipper } from "igniteui-react-core";
import { ChartSelection } from "./ChartSelection";
import { PropertyMetadata } from "igniteui-react-core";
import { round10N } from "igniteui-react-core";
/**
 * @hidden
 */
var CategoryBucketCalculator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryBucketCalculator, _super);
    function CategoryBucketCalculator(a) {
        var _this = _super.call(this) || this;
        _this.c = null;
        _this.e = 0;
        _this.f = 0;
        _this.d = 0;
        if (a == null) {
            throw new ArgumentNullException(0, "view");
        }
        _this.c = a;
        _this.e = -1;
        _this.d = 0;
        return _this;
    }
    CategoryBucketCalculator.prototype.getBucket = function (a) {
        return null;
    };
    CategoryBucketCalculator.prototype.getBucketWithoutUnknowns = function (a) {
        return this.getBucket(a);
    };
    CategoryBucketCalculator.prototype.getErrorBucket = function (a, b) {
        return NaN;
    };
    CategoryBucketCalculator.prototype.g = function (a) {
        var b = this.c.by;
        var c = this.c.bx;
        var d = this.c.cy;
        var e = typeCast(CategoryAxisBase.$, d.fetchXAxis());
        var f = d.dn;
        if (b.isEmpty || c.isEmpty || e == null || f == null || f.count == 0) {
            this.d = 0;
            return;
        }
        var g = d.getEffectiveViewport1(this.c);
        var h = d.wj(this.c);
        var i = typeCast(ISortingAxis_$type, e);
        if (i == null || i.i$b == null) {
            var j = new ScalerParams(0, b, c, e.ch, g);
            var k = Math.floor(e.eo(h.left, j));
            var l = Math.ceil(e.eo(h.right, j));
            if (e.ch) {
                l = Math.ceil(e.eo(h.left, j));
                k = Math.floor(e.eo(h.right, j));
            }
            l = Math.min(l, e.lk - 1);
            var m = Math.floor((l - k + 1) * a / h.width);
            this.d = truncate(Math.max(1, m));
            this.e = truncate(Math.max(0, Math.floor(k / this.d) - 1));
            this.f = truncate(Math.ceil(l / this.d));
        }
        else {
            this.e = i.i$e(b, c, g, h);
            this.f = i.i$g(b, c, g, h);
            this.d = 1;
        }
    };
    CategoryBucketCalculator.prototype.getBucketInfo = function (a, b, c, d) {
        a = this.e;
        b = this.f;
        c = this.d;
        d = this.c.cy.id;
        return {
            p0: a,
            p1: b,
            p2: c,
            p3: d
        };
    };
    CategoryBucketCalculator.prototype.cacheValues = function () {
    };
    CategoryBucketCalculator.prototype.unCacheValues = function () {
    };
    CategoryBucketCalculator.$t = markType(CategoryBucketCalculator, 'CategoryBucketCalculator', Base.$, [IBucketizer_$type]);
    return CategoryBucketCalculator;
}(Base));
export { CategoryBucketCalculator };
/**
 * @hidden
 */
var CategoryFramePreparerBase = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryFramePreparerBase, _super);
    function CategoryFramePreparerBase(a) {
        var _rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            _rest[_i - 1] = arguments[_i];
        }
        var _this = this;
        a = (a == void 0) ? 0 : a;
        switch (a) {
            case 0:
                {
                    var c = _rest[0];
                    var d = [c, typeCast(ISupportsMarkers_$type, c), typeCast(IProvidesViewport_$type, c), typeCast(ISupportsErrorBars_$type, c), typeCast(IBucketizer_$type, c)];
                    {
                        var f = d[0];
                        var g = d[1];
                        var h = d[2];
                        var i = d[3];
                        var j = d[4];
                        _this = _super.call(this, 1, g, h, i) || this;
                        _this.g = null;
                        _this.h = null;
                        _this.m = false;
                        _this.o = new List$1(Array_$type, 0);
                        _this.p = new List$1(Number_$type, 0);
                        _this.n = new List$1(Number_$type, 0);
                        _this.h = f;
                        _this.g = j;
                    }
                }
                break;
            case 1:
                {
                    var c = _rest[0];
                    var d = _rest[1];
                    var e = _rest[2];
                    var f = _rest[3];
                    var g = _rest[4];
                    _this = _super.call(this, 1, d, e, f) || this;
                    _this.g = null;
                    _this.h = null;
                    _this.m = false;
                    _this.o = new List$1(Array_$type, 0);
                    _this.p = new List$1(Number_$type, 0);
                    _this.n = new List$1(Number_$type, 0);
                    _this.h = c;
                    _this.g = g;
                }
                break;
        }
        return _this;
    }
    CategoryFramePreparerBase.prototype.i = function (a) {
        var b = new PreparationParams();
        b.f = this.h.useHighMarkerFidelity;
        b.b = this.h.scaler;
        b.c = this.h.yScaler;
        b.d = typeCast(ISortingAxis_$type, b.b);
        b.a = typeCast(CategoryFrame.$, a);
        if (b.a == null || b.b == null || b.c == null) {
            return null;
        }
        var c;
        var d;
        var e;
        var f;
        var g = this.g.getBucketInfo(c, d, e, f);
        c = g.p0;
        d = g.p1;
        e = g.p2;
        f = g.p3;
        b.k = c;
        b.l = d;
        b.j = e;
        b.i = f;
        if (b.l < b.k) {
            return null;
        }
        var h;
        var i;
        var j;
        var k = this.a.getViewInfo(i, h);
        i = k.p0;
        h = k.p1;
        j = this.a.getEffectiveViewport();
        var l = this.a.getMainContentViewport();
        b.p = h;
        b.o = i;
        b.n = j;
        b.m = l;
        if (Rect.l_op_Equality(b.p, Rect.empty) || Rect.l_op_Equality(b.o, Rect.empty)) {
            return null;
        }
        if (this.h != null && (typeCast(Series.$, this.h) !== null && this.h.fb) && this.g != null) {
            b.e = true;
        }
        return b;
    };
    CategoryFramePreparerBase.prototype.r = function (a, b, c, d) {
        return a.getSeriesOffset(b, c, d, this.h.currentMode2Index, this.h.currentCategoryMode);
    };
    CategoryFramePreparerBase.prototype.s = function (a) {
        return this.r(a.b, a.p, a.o, a.n);
    };
    CategoryFramePreparerBase.prototype.d = function (a, b) {
        var c = this.i(a);
        if (c == null || this.g == null) {
            return;
        }
        c.a.f.clear();
        c.a.k.clear();
        c.a.n.clear();
        c.a.q.clear();
        c.a.i.clear();
        c.a.h.clear();
        c.a.g.clear();
        if (c.a.m != null) {
            c.a.m.clear();
        }
        var d = this.c.shouldDisplayMarkers;
        var e = false;
        var f = this.s(c);
        var g = this.j(c);
        if (c.d != null && c.d.i$b != null && c.d.i$b.count != g.e) {
            return;
        }
        if (c.d != null && c.d.i$a) {
            c.d.i$h();
        }
        this.g.cacheValues();
        if (this.m && c.d != null && c.d.i$b.count > 0) {
            var h = c.d.i$b._inner[c.k];
            this.z(g, h, true, c.e);
            while ((isNaN_(g.c) || isNaN_(g.d)) && c.k > 0) {
                c.k = c.k - 1;
                h = c.d.i$b._inner[c.k];
                this.z(g, h, true, c.e);
            }
            var i = c.d.i$b._inner[c.l];
            this.z(g, i, true, c.e);
            while ((isNaN_(g.c) || isNaN_(g.d)) && c.l < g.e - 1) {
                c.l = c.l + 1;
                i = c.d.i$b._inner[c.l];
                this.z(g, i, true, c.e);
            }
        }
        if (this.m && c.d == null) {
            var j = c.k;
            var k = this.g.getBucket(j);
            while ((isNaN_(k[1]) || isNaN_(k[2])) && c.k > 0) {
                c.k = c.k - 1;
                j = c.k;
                k = this.g.getBucket(j);
            }
            var l = c.l;
            k = this.g.getBucket(l);
            while ((isNaN_(k[1]) || isNaN_(k[2])) && c.l < (intDivide(g.e, c.j))) {
                c.l = c.l + 1;
                l = c.l;
                k = this.g.getBucket(l);
            }
        }
        this.w(c, g, f);
        var m = this.t(c, g, f, d, e);
        this.c.updateMarkerCount(m);
        this.v(typeCast(CategoryFrame.$, a), b);
        this.g.unCacheValues();
        return;
    };
    CategoryFramePreparerBase.prototype.v = function (a, b) {
    };
    Object.defineProperty(CategoryFramePreparerBase.prototype, "q", {
        get: function () {
            return this.o;
        },
        enumerable: false,
        configurable: true
    });
    CategoryFramePreparerBase.prototype.t = function (a, b, c, d, e) {
        var _this = this;
        var f = 0;
        var g = false;
        var h = null;
        var i = false;
        var j = a.d != null;
        var k = a.p;
        var l = a.o;
        var m = typeCast(NumericAxisBase.$, a.c) !== null && a.c.lp;
        var n = a.f;
        var o = this.h.mayContainUnknowns;
        var p = this.o;
        var q = this.n;
        var r = this.p;
        p.clear();
        q.clear();
        r.clear();
        var s = this.g;
        var t = this.h.provideCollisionDetector();
        var u = 0;
        var v = a.n;
        var w = new ScalerParams(0, k, l, a.b.j$b, v);
        var x = ((function () {
            var $ret = new ScalerParams(0, k, l, a.c.j$b, v);
            $ret.b = typeCast(Series.$, _this.h) !== null ? _this.h.iy() : NaN;
            return $ret;
        })());
        if (j) {
            u = a.b.j$d(2, w) - a.b.j$d(1, w);
            i = this.k();
        }
        var y = a.k;
        var z = a.l;
        var aa = a.a.f;
        var ab = typeCast(IDateTimeAxis_$type, a.b);
        if (ab != null) {
            ab.StartTimeCursor();
        }
        var _loop_1 = function (ac) {
            var ad = void 0;
            if (!j) {
                if (!o) {
                    ad = s.getBucketWithoutUnknowns(ac);
                }
                else {
                    ad = s.getBucket(ac);
                }
            }
            else {
                ad = ((function () { var ae = _this.f(a, ac, b, u, g, h, c); ac = ae.p1; g = ae.p4; h = ae.p5; return ae.ret; })());
            }
            var ae = !m || (m && ad[1] > 0);
            var af = ad[0];
            if (!j || !isNaN_(ad[0])) {
                aa.add(ad);
                if (j) {
                    if (a.a.m == null) {
                        a.a.m = new List$1(Number_$type, 0);
                    }
                    a.a.m.add(ac * a.j);
                }
                if (g && i) {
                    if (h != null) {
                        aa.add(h);
                        if (j) {
                            if (a.a.m == null) {
                                a.a.m = new List$1(Number_$type, 0);
                            }
                            a.a.m.add(ac * a.j);
                        }
                    }
                }
                if (d && ae) {
                    var ag = ac * a.j;
                    var ah = ag;
                    if (j && a.d.i$b != null && ag >= 0 && ag < a.d.i$b.count) {
                        ag = a.d.i$b._inner[ag];
                    }
                    var ai = new Array(ad.length);
                    ai[0] = ad[0];
                    ai[1] = ad[1];
                    ai[2] = ad[2];
                    if (n && ag < b.e) {
                        ai = new Array(ad.length);
                        ai[0] = af;
                        this_1.z(b, ag, true, a.e);
                        ai[1] = b.c;
                        ai[2] = b.d;
                    }
                    p.add(ai);
                    q.add(a.a.f.count - 1);
                    r.add(ag);
                }
            }
            out_ac_1 = ac;
        };
        var this_1 = this, out_ac_1;
        for (var ac = y; ac <= z; ++ac) {
            _loop_1(ac);
            ac = out_ac_1;
        }
        if (ab != null) {
            ab.ResetTimeCursor();
        }
        this.y(a, c, j, w, x, e, d);
        if (d) {
            var aj = p.count;
            var ak = void 0;
            for (var al = 0; al < aj; al++) {
                ak = p._inner[al];
                if (this.l(a.a, ak, t, Math.min(r._inner[al], b.e - 1), f, q._inner[al])) {
                    ++f;
                }
            }
        }
        return f;
    };
    CategoryFramePreparerBase.prototype.k = function () {
        if (typeCast(Series.$, this.h) !== null) {
            return this.h.ev;
        }
        return false;
    };
    CategoryFramePreparerBase.prototype.f = function (a, b, c, d, e, f, g) {
        var h;
        f = null;
        e = false;
        var i = typeCast(CategorySeries.$, this.h);
        var j = a.o;
        var k = a.p;
        var l = a.n;
        var m = new ScalerParams(0, k, j, a.b.j$b, l);
        var n = new ScalerParams(0, k, j, a.c.j$b, l);
        var o = a.e;
        if (a.d == null || a.d.i$b == null || a.d.i$b.count == 0) {
            return {
                ret: [NaN, NaN, NaN],
                p1: b,
                p4: e,
                p5: f
            };
        }
        if (i != null && i.dn != null && i.dn.count < a.d.i$b.count) {
            return {
                ret: [NaN, NaN, NaN],
                p1: b,
                p4: e,
                p5: f
            };
        }
        var p = a.d.i$b._inner[b];
        var q = a.d.i$d(p);
        var r = q;
        this.z(c, p, false, o);
        while (b < a.l) {
            p = a.d.i$b._inner[b + 1];
            r = a.d.i$d(p);
            this.z(c, p, true, o);
            if (r - q >= d || isNaN_(c.c) || isNaN_(c.d)) {
                if (e) {
                    var s = a.d.i$b._inner[b];
                    this.z(c, s, true, o);
                    f = [a.b.j$c(q, m), c.c, c.d];
                }
                break;
            }
            if (!e && this.k()) {
                var t = a.d.i$b._inner[b];
                this.z(c, t, true, o);
                var u = [a.b.j$c(q, m), c.c, c.d];
                if (!isNaN_(u[0])) {
                    if (!isNaN_(u[1]) && !isNaN_(u[2])) {
                        if (a.d != null) {
                            u[0] = u[0] + g;
                        }
                        a.a.f.add(u);
                        if (a.d != null) {
                            if (a.a.m == null) {
                                a.a.m = new List$1(Number_$type, 0);
                            }
                            a.a.m.add(b);
                        }
                        e = true;
                    }
                    else {
                        break;
                    }
                }
            }
            b++;
            this.u(c, p, o);
        }
        var v = NaN;
        if (!isNaN_(q)) {
            v = a.b.j$c(q, m);
        }
        h = this.e(v, c);
        return {
            ret: h,
            p1: b,
            p4: e,
            p5: f
        };
    };
    CategoryFramePreparerBase.prototype.y = function (a, b, c, d, e, f, g) {
        var h = a.j;
        var i = a.a.f.count;
        var j = a.a.f;
        var k = this.o;
        var l = k.count;
        if (f) {
            var m = a.c.j$c(a.c.me, e);
            a.c.j$f(a.a.k, 0, a.a.k.count, e);
            var n = a.a.k;
            var o = n.count;
            for (var p = 0; p < n.count; p++) {
                n._inner[p] = Math.abs(n._inner[p] - m);
            }
        }
    };
    CategoryFramePreparerBase.prototype.w = function (a, b, c) {
    };
    CategoryFramePreparerBase.$t = markType(CategoryFramePreparerBase, 'CategoryFramePreparerBase', FramePreparer.$);
    return CategoryFramePreparerBase;
}(FramePreparer));
export { CategoryFramePreparerBase };
/**
 * @hidden
 */
var CategoryFramePreparer = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryFramePreparer, _super);
    function CategoryFramePreparer(a) {
        var _rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            _rest[_i - 1] = arguments[_i];
        }
        var _this = this;
        a = (a == void 0) ? 0 : a;
        switch (a) {
            case 0:
                {
                    var c = _rest[0];
                    var d = [c, typeCast(ISupportsMarkers_$type, c), typeCast(IProvidesViewport_$type, c), typeCast(ISupportsErrorBars_$type, c), typeCast(IBucketizer_$type, c)];
                    {
                        var f = d[0];
                        var g = d[1];
                        var h = d[2];
                        var i = d[3];
                        var j = d[4];
                        _this = _super.call(this, 1, f, g, h, i, j) || this;
                        _this.aa = new DefaultCategoryTrendlineHost();
                        _this.ab = new DefaultSingleValueProvider();
                        _this.ac = null;
                        _this.ad = null;
                        _this.ac = _this.aa;
                        if (typeCast(IHasCategoryTrendline_$type, f) !== null) {
                            _this.ac = typeCast(IHasCategoryTrendline_$type, f);
                        }
                        _this.ad = _this.ab;
                        if (typeCast(IHasSingleValueCategory_$type, f) !== null) {
                            _this.ad = typeCast(IHasSingleValueCategory_$type, f);
                        }
                    }
                }
                break;
            case 1:
                {
                    var c = _rest[0];
                    var d = _rest[1];
                    var e = _rest[2];
                    var f = _rest[3];
                    var g = _rest[4];
                    _this = _super.call(this, 1, c, d, e, f, g) || this;
                    _this.aa = new DefaultCategoryTrendlineHost();
                    _this.ab = new DefaultSingleValueProvider();
                    _this.ac = null;
                    _this.ad = null;
                    _this.ac = _this.aa;
                    if (typeCast(IHasCategoryTrendline_$type, c) !== null) {
                        _this.ac = typeCast(IHasCategoryTrendline_$type, c);
                    }
                    _this.ad = _this.ab;
                    if (typeCast(IHasSingleValueCategory_$type, c) !== null) {
                        _this.ad = typeCast(IHasSingleValueCategory_$type, c);
                    }
                }
                break;
        }
        return _this;
    }
    CategoryFramePreparer.prototype.l = function (a, b, c, d, e, f) {
        var g = b[0];
        var h = b[1];
        var i = new Rect(0, g - 5, h - 5, 11, 11);
        if (!isNaN_(g) && !isNaN_(h) && !isInfinity(g) && !isInfinity(h) && c.tryAdd(i)) {
            a.n.add({ $type: Point_$type, x: g, y: h });
            this.c.updateMarkerTemplate(e, d, f);
            return true;
        }
        return false;
    };
    CategoryFramePreparer.prototype.j = function (a) {
        var b = new SingleValuesHolder();
        var c = this.ad.valueColumn;
        b.f = c;
        return b;
    };
    CategoryFramePreparer.prototype.x = function (a, b, c, d, e, f) {
        if (d) {
            b[0] = b[0] + c;
        }
        else {
            b[0] = a.b.j$c(b[0], e) + c;
        }
        b[1] = a.c.j$c(b[1], f);
        if (a.j > 1 || d) {
            b[2] = a.c.j$c(b[2], f);
        }
        else {
            b[2] = b[1];
        }
    };
    CategoryFramePreparer.prototype.y = function (a, b, c, d, e, f, g) {
        _super.prototype.y.call(this, a, b, c, d, e, f, g);
        var h = a.j;
        var i = a.a.f.count;
        var j = a.a.f;
        var k = this.q;
        var l = k.count;
        var m = new List$1(Number_$type, 0);
        if (!c) {
            m.add(0);
        }
        var n = new List$1(Number_$type, 0);
        n.add(1);
        if (c || h > 1) {
            n.add(2);
        }
        a.b.j$e(j, m, 0, i, d);
        a.c.j$e(j, n, 0, i, e);
        for (var o = 0; o < i; o++) {
            j._inner[o][0] = j._inner[o][0] + b;
        }
        if (h == 1 && !c) {
            for (var p = 0; p < i; p++) {
                j._inner[p][2] = j._inner[p][1];
            }
        }
        if (g) {
            a.b.j$e(k, m, 0, l, d);
            a.c.j$e(k, n, 0, l, e);
            for (var q = 0; q < l; q++) {
                k._inner[q][0] = k._inner[q][0] + b;
            }
            if (h == 1 && !c) {
                for (var r = 0; r < l; r++) {
                    k._inner[r][2] = k._inner[r][1];
                }
            }
        }
    };
    CategoryFramePreparer.prototype.w = function (a, b, c) {
        if (this.ac.trendLineType == 0 || this.ac.trendlinePreparer == null || this.ac.trendLinePeriod < 1) {
            return;
        }
        this.ac.trendlinePreparer.prepareLine(this.ac, a, b, c);
    };
    CategoryFramePreparer.prototype.z = function (a, b, c, d) {
        var e = a;
        var f = e.f;
        var g = f.item(b);
        var h = g;
        if (d) {
            var i = this.g.getBucket(b);
            g = i[1];
            h = i[1];
        }
        if (c) {
            e.c = g;
            e.d = h;
        }
        else {
            e.a = g;
            e.b = h;
        }
    };
    CategoryFramePreparer.prototype.u = function (a, b, c) {
        var d = a;
        var e = d.f;
        if (b < e.count) {
            var f = void 0;
            if (c) {
                var g = this.g.getBucket(b);
                f = g[1];
            }
            else {
                f = e.item(b);
            }
            d.a = Math.min(d.a, f);
            d.b = Math.max(d.b, f);
        }
    };
    CategoryFramePreparer.prototype.e = function (a, b) {
        var c = b;
        return [a, c.a, c.b];
    };
    CategoryFramePreparer.$t = markType(CategoryFramePreparer, 'CategoryFramePreparer', CategoryFramePreparerBase.$);
    return CategoryFramePreparer;
}(CategoryFramePreparerBase));
export { CategoryFramePreparer };
/**
 * @hidden
 */
var CategorySeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategorySeries, _super);
    function CategorySeries() {
        var _this = _super.call(this) || this;
        _this.yz = null;
        _this.y7 = null;
        _this.zg = false;
        _this.zf = false;
        _this.assigningCategoryStyle = null;
        _this.assigningCategoryMarkerStyle = null;
        _this.zd = null;
        _this.ze = null;
        _this.yw = new CategoryFrame(3);
        _this.yy = new CategoryFrame(3);
        _this.yu = new CategoryFrame(3);
        _this.yx = new CategoryFrame(3);
        _this.y2 = null;
        _this.y6 = null;
        _this.y2 = new CategorySeriesRenderManager();
        _this.y6 = new CategoryTransitionSourceFramePreparer();
        _this.zg = false;
        _this.d2 = false;
        _this.zf = false;
        _this.ab = CategorySeries.$;
        return _this;
    }
    CategorySeries.prototype.cs = function () {
        var a = new CategorySeriesView(this);
        return a;
    };
    Object.defineProperty(CategorySeries.prototype, "y8", {
        get: function () {
            var _this = this;
            if (this.y7 == null) {
                this.y7 = DVContainer.instance.createInstance(ICategorySeriesInteractionManager_$type, function () { return _this.dr.n; });
            }
            return this.y7;
        },
        enumerable: false,
        configurable: true
    });
    CategorySeries.prototype.eb = function (a, b, c, d, e) {
        var _this = this;
        if (!this.hasMarkers) {
            return false;
        }
        var f = this.getEffectiveViewport1(this.cv);
        var g = this.ys();
        var h = new ScalerParams(1, this.db.v4, this.ws, g.ch);
        h.c = f;
        var i = this.y9();
        var j = new ScalerParams(1, this.db.v4, this.ws, i.ch);
        j.c = f;
        var k = this.db.v4;
        var l = this.ws;
        var m = this.currentMode2Index;
        var n = 0;
        if (g.ce && g.cc) {
            n = g.getSeriesOffset(k, l, f, m, g.categoryMode);
        }
        a.clear();
        this.w6.cj(function (o) {
            if (o._visibility == 0) {
                var p = new SeriesPointOfInterest();
                var q = o.n;
                var r = o.o;
                if (q < _this.ws.left || q > _this.ws.right || r < _this.ws.top || r > _this.ws.bottom) {
                    return;
                }
                if (_this.isVertical) {
                    if (g.cb && !g.ce && g.categoryMode != 0) {
                        p.e = o.ak;
                    }
                    else {
                        p.e = g.eo(r, h);
                    }
                    p.d = i.eo(q, j);
                }
                else {
                    if (g.cb && !g.ce && g.categoryMode != 0) {
                        p.d = o.ak;
                    }
                    else {
                        p.d = g.eo(q, h);
                    }
                    p.e = i.eo(r, j);
                }
                p.b = o.content;
                if (b) {
                    if (_this.isVertical) {
                        if (c) {
                            p.f = _this.zz(p.d, d);
                        }
                        else {
                            var s = _this.jd(o);
                            if (!isNaN_(s)) {
                                p.f = _this.zz(s, d);
                            }
                            else {
                                p.f = _this.zz(p.d, d);
                            }
                        }
                    }
                    else {
                        if (c) {
                            p.f = _this.zz(p.e, d);
                        }
                        else {
                            var t = _this.jd(o);
                            if (!isNaN_(t)) {
                                p.f = _this.zz(t, d);
                            }
                            else {
                                p.f = _this.zz(p.e, d);
                            }
                        }
                    }
                }
                else {
                    var u = _this.au(true);
                    if (u != null) {
                        var v = _this.ks(o.ak, false);
                        if (v != null) {
                            p.f = u.i(v);
                        }
                        else {
                            p.f = null;
                        }
                    }
                    else {
                        p.f = g.gf(o.ak);
                    }
                }
                p.c = o.ak;
                p.g = _this.vg;
                if (_this.g9) {
                    p.g = _this.vn(o.ak);
                }
                a.add(p);
            }
        });
        return true;
    };
    CategorySeries.prototype.zz = function (a, b) {
        var c;
        if (b >= 0) {
            c = round10N(a, b);
        }
        else {
            c = a;
        }
        var d = this.au(false);
        if (d != null) {
            var e = d;
            return e.i(c);
        }
        else {
            return c.toString();
        }
    };
    CategorySeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.y3 = a;
    };
    CategorySeries.prototype.preferredCategoryMode = function (a) {
        return 0;
    };
    Object.defineProperty(CategorySeries.prototype, "categoryAxis", {
        get: function () {
            return this.ys();
        },
        enumerable: false,
        configurable: true
    });
    CategorySeries.prototype.ys = function () {
        return this.fetchXAxis();
    };
    CategorySeries.prototype.y9 = function () {
        return this.fetchYAxis();
    };
    CategorySeries.prototype.get_ek = function () {
        return true;
    };
    CategorySeries.prototype.get_f5 = function () {
        return this.y9() != null && this.y9().ch;
    };
    CategorySeries.prototype.get_ey = function () {
        return true;
    };
    CategorySeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.y3.cw.g(this.id);
    };
    CategorySeries.prototype.zm = function () {
        return (this.assigningCategoryStyle != null && this.zg) || (this.bx != 1) || this.d2 || (this.cp != 1) || (this.co != 1);
    };
    CategorySeries.prototype.zn = function () {
        return (this.assigningCategoryMarkerStyle != null && this.zf) || (this.bx != 1) || this.d2 || (this.cp != 1) || (this.co != 1);
    };
    CategorySeries.prototype.aaj = function (a) {
        if (this.assigningCategoryStyle != null && this.zg) {
            this.assigningCategoryStyle(this, a);
        }
    };
    CategorySeries.prototype.aai = function (a) {
        if (this.assigningCategoryMarkerStyle != null && this.zf) {
            this.assigningCategoryMarkerStyle(this, a);
        }
    };
    CategorySeries.prototype.ym = function (a, b) {
        return null;
    };
    CategorySeries.prototype.o5 = function (a, b) {
        _super.prototype.o5.call(this, a, b);
        if (a) {
            this.z8(b);
        }
    };
    CategorySeries.prototype.br = function () {
        var a = _super.prototype.br.call(this);
        return a;
    };
    CategorySeries.prototype.qb = function () {
        _super.prototype.qb.call(this);
        var a = this.fetchXAxis();
        if (a != null) {
            a.ib(false);
        }
        var b = this.fetchYAxis();
        if (b != null) {
            b.ib(false);
        }
    };
    CategorySeries.prototype.sr = function (a, b) {
        _super.prototype.sr.call(this, a, b);
        this.y3.cw.g(this.id);
        this.rr(false);
    };
    CategorySeries.prototype.sq = function (a, b) {
        if (this.g1 && this.jq < 0.05 && (this.db == null || !this.db.fn())) {
            this.g2 = true;
            this.g5 = false;
            this.gk = false;
        }
        _super.prototype.sq.call(this, a, b);
        this.y3.cw.g(this.id);
        this.rr(false);
    };
    CategorySeries.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case "SeriesViewer":
                var e = this.fetchYAxis();
                var f = this.fetchXAxis();
                if (c != null && d == null) {
                    this.pc(f);
                    this.pc(e);
                }
                if (c == null && d != null) {
                    this.rh(f);
                    this.rh(e);
                }
                this.y3.cw.g(this.id);
                this.rr(false);
                break;
            case "SyncLink":
                if (this.c3 != null && this.db != null) {
                    this.y3.cw.g(this.id);
                }
                this.rr(false);
                break;
            case "TransitionProgress":
                this.yy.c(this.jq, this.yw, this.yu);
                if (this.d5(this.cv)) {
                    return;
                }
                if (this.jq == 1) {
                    this.yu.w();
                    this.aak(this.yu, this.y3);
                    if (this.g1) {
                        this.g1 = false;
                        this.o6();
                        this.dc.r = truncate(this.i9());
                        this.dc.b = this.dd;
                    }
                    if (this.g4) {
                        this.g4 = false;
                        this.o6();
                        this.dc.r = truncate(this.i9());
                        this.dc.b = this.dd;
                        this.q0();
                    }
                    this.qo();
                }
                else {
                    this.aak(this.yy, this.y3);
                }
                if (this.db != null) {
                    this.db.qo(this);
                }
                break;
            case "ErrorBarSettings":
                this.ql();
                break;
            case CategorySeries.$$p[5]:
                this.rr(false);
                break;
            case CategorySeries.$$p[2]:
                this.zg = this.zi();
                this.rr(false);
                break;
            case CategorySeries.$$p[1]:
                this.zf = this.zh();
                this.rr(false);
                break;
        }
    };
    CategorySeries.prototype.zi = function () {
        return this.zk;
    };
    CategorySeries.prototype.zh = function () {
        return this.zj;
    };
    CategorySeries.prototype.zt = function (a) {
        return this.it(a, this.fetchXAxis());
    };
    CategorySeries.prototype.get_fs = function () {
        return typeCast(ISortingAxis_$type, this.ys()) !== null;
    };
    CategorySeries.prototype.a0 = function () {
        return this.ys();
    };
    CategorySeries.prototype.j5 = function (a) {
        return this.y8.b$d(this, a);
    };
    CategorySeries.prototype.kn = function (a) {
        return this.y8.b$f(this, a);
    };
    CategorySeries.prototype.j3 = function (a) {
        var b = truncate(Math.round(this.ir(a)));
        return b;
    };
    CategorySeries.prototype.ir = function (a) {
        return this.y8.b$c(this, a);
    };
    CategorySeries.prototype.dh = function (a, b, c) {
        if (this.fz(this.bx) && !c) {
            var d = this.zc(a, b, this.ze);
            this.ze = d;
            return d;
        }
        else {
            var e = this.za(a, b, this.ze);
            e.d = true;
            this.ze = e;
            return e;
        }
    };
    CategorySeries.prototype.get_fe = function () {
        return true;
    };
    CategorySeries.prototype.o3 = function () {
        _super.prototype.o3.call(this);
        this.zd = null;
        this.ze = null;
    };
    CategorySeries.prototype.dg = function (a, b, c) {
        if (this.ed && this.fz(this.bx) && !c) {
            var d = this.zb(a, b, this.zd);
            this.zd = d;
            return d;
        }
        else {
            var e = this.za(a, b, this.zd);
            this.zd = e;
            return e;
        }
    };
    CategorySeries.prototype.zs = function () {
        return this.y3.cz.a;
    };
    CategorySeries.prototype.zc = function (a, b, c) {
        if (!this.xg()) {
            return null;
        }
        var d = this.yv();
        var e = this.zs();
        return this.y8.b$b(this, e, this.zw(this.cv), this.zv(this.cv), d.f, a, b, c);
    };
    CategorySeries.prototype.zb = function (a, b, c) {
        var d = this.zw(this.cv);
        var e = this.zx(this.cv);
        var f = this.zv(this.cv);
        var g = this.yv();
        return this.y8.b$a(this, g.f, g.m, d, e, f, a, b, c);
    };
    CategorySeries.prototype.za = function (a, b, c) {
        var d = new HighlightingInfo();
        d.i = this;
        d.g = 0;
        d.f = this.dn != null ? this.dn.count - 1 : 0;
        d.c = true;
        if (c != null && c.g == d.g && c.f == d.f) {
            return c;
        }
        return d;
    };
    CategorySeries.prototype.yv = function () {
        var a = this.yu;
        if (this.d3()) {
            if (this.dc.g) {
                this.dc.t();
            }
            a = this.yy;
        }
        return a;
    };
    CategorySeries.prototype.aaf = function (a, b, c, d, e, f, g) {
        var h = this.bx != 1;
        var i = this.d2;
        var j = this.cp != 1;
        var k = this.co != 1;
        this.y2.a3(this.y2.d, d.cn, a, b, c, d, e, f, true, g);
        this.y2.aa(this.ib);
        var l = this.y2.ax;
        var m = null;
        var n = this.dn;
        var o = null;
        var p = null;
        var q = null;
        var r = null;
        var s = null;
        if (i && this.db != null && !l.e) {
            var t = this.vo(l.z);
            var u = this.vp(l.z);
            this.y2.ae = t;
            this.y2.af = u;
            l.aa = t;
            l.ab = u;
        }
        if (h && this.db != null) {
            m = this.db.dl.getHighlightingInfo(this, n, l.z, l.y, true);
            if (this.db.dl.isCrossContextHighlight(this)) {
                o = this.db.dl.getHighlightingInfo(this, n, l.z, l.y, true);
            }
            l.c = m;
        }
        if (j && this.db != null) {
            p = this.db.dm.getHighlightingInfo(this, n, l.z, l.y, true);
            if (p == null && this.bn != null) {
                p = this.db.dm.getHighlightingInfo(this.bn, this.bn.dn, l.z, l.y, true);
            }
            if (this.db.dm.isCrossContextHighlight(this)) {
                r = this.db.dm.getHighlightingInfo(this, n, l.z, l.y, true);
            }
            l.d = p;
        }
        if (k && this.db != null) {
            q = this.db.dk.getHighlightingInfo(this, n, l.z, l.y, true);
            if (this.db.dk.isCrossContextHighlight(this)) {
                s = this.db.dk.getHighlightingInfo(this, n, l.z, l.y, true);
            }
            l.b = q;
        }
        l.h = f;
        this.aai(l);
        this.y2.y(this, m, p, q, f, h, j, k, l.v, this.db.dl.totalHighWaterMark, l.x, this.db.dm.totalHighWaterMark, l.t, this.db.dk.totalHighWaterMark);
    };
    CategorySeries.prototype.aag = function (a, b, c, d, e, f, g) {
        var h = this.bx != 1;
        var i = this.cp != 1;
        var j = this.co != 1;
        var k = this.d2;
        this.y2.a3(this.y2.b, d.cn, a, b, c, d, e, f, false, g);
        this.y2.ab(this.ib);
        var l = this.y2.ay;
        var m = null;
        var n = null;
        var o = null;
        var p = this.dn;
        if (k && this.db != null && !l.e) {
            var q = this.vn(l.z);
            var r = this.vq(l.z);
            this.y2.ag = q;
            this.y2.ah = r;
            l.aa = q;
            l.ab = r;
        }
        if (h && this.db != null) {
            m = this.db.dl.getHighlightingInfo(this, p, l.z, l.y, false);
            l.c = m;
        }
        if (i && this.db != null) {
            n = this.db.dm.getHighlightingInfo(this, p, l.z, l.y, false);
            if (n == null && this.bn != null) {
                n = this.db.dm.getHighlightingInfo(this.bn, this.bn.dn, l.z, l.y, false);
            }
            l.d = n;
        }
        if (j && this.db != null) {
            o = this.db.dk.getHighlightingInfo(this, p, l.z, l.y, false);
            if (o == null && this.bn != null) {
                o = this.db.dk.getHighlightingInfo(this.bn, this.bn.dn, l.z, l.y, false);
            }
            l.b = o;
        }
        l.h = f;
        this.aaj(l);
        this.y2.z(this, m, n, o, f, h, i, j, b < 0);
    };
    CategorySeries.prototype.he = function (a, b, c) {
        var d = _super.prototype.he.call(this, a, b, c);
        d = d && this.zr(this.fetchXAxis());
        d = d && this.zr(this.fetchYAxis());
        var e = c;
        if (!c.o() || b.isEmpty || a.isEmpty || this.dn == null) {
            d = false;
        }
        if (!d) {
            e.cw.d = 0;
        }
        return d;
    };
    CategorySeries.prototype.zr = function (a) {
        if (a == null || a.bi == null) {
            return false;
        }
        var b = typeCast(CategoryAxisBase.$, a);
        if (b != null) {
            if (b.itemsSource == null) {
                return false;
            }
            if (b.lj < 1) {
                return false;
            }
        }
        else {
            var c = typeCast(NumericAxisBase.$, a);
            if (c != null) {
                return c.l4 != c.l3;
            }
        }
        return true;
    };
    CategorySeries.prototype.gn = function () {
        return this.zl;
    };
    CategorySeries.prototype.go = function () {
        return true;
    };
    CategorySeries.prototype.y4 = function () {
        return 12;
    };
    CategorySeries.prototype.o6 = function () {
        _super.prototype.o6.call(this);
        this.yw.w();
        this.yu.w();
        this.yy.w();
    };
    CategorySeries.prototype.rs = function (a) {
        _super.prototype.rs.call(this, a);
        this.y3.cw.g(this.id);
        if (this.d5(this.cv)) {
            this.yu.z();
            this.yy.z();
            this.yw.z();
            return;
        }
        if (this.gr) {
            if (this.d3()) {
                if (this.dc.g) {
                    this.dc.t();
                }
                this.aak(this.yy, this.y3);
            }
            else {
                this.aak(this.yu, this.y3);
            }
            return;
        }
        if (this.gf(a)) {
            var b = this.yw;
            if (this.d3()) {
                if (this.dc.g) {
                    this.dc.t();
                }
                this.yw = this.yy;
                this.yy = b;
            }
            else {
                this.yw = this.yu;
                this.yu = b;
            }
            var c = false;
            this.aah(this.yu, this.y3);
            if (this.g3) {
                this.dc.w();
                this.dc.r = truncate(this.ja());
                this.dc.b = this.de != null ? this.de : this.dd;
                if (this.g4) {
                    var d = this.yy;
                    this.yy = this.yw;
                    this.yw = d;
                    c = true;
                }
                this.y6.s(this, this.yw, this.yu, this.isVertical, this.fetchXAxis(), this.fetchYAxis(), this.y5, this.y4(), this.c5, this.c4(), runOn(this, this.zu), this.y3.bx);
                if (this.g4) {
                    var e = this.yy;
                    this.yy = this.yw;
                    this.yw = e;
                }
                this.g4 = false;
            }
            if (this.g6) {
                this.dc.w();
                this.dc.r = truncate(this.jb());
                this.dc.b = this.df != null ? this.df : this.dd;
                if (this.g1) {
                    var f = this.yy;
                    this.yy = this.yw;
                    this.yw = f;
                    c = true;
                }
                this.y6.s(this, this.yw, this.yu, this.isVertical, this.fetchXAxis(), this.fetchYAxis(), this.y5, this.y4(), this.c5, this.c4(), runOn(this, this.zu), this.y3.bx);
                var g = this.yu;
                this.yu = this.yw;
                this.yw = g;
                if (this.g1) {
                    var h = this.yy;
                    this.yy = this.yw;
                    this.yw = h;
                }
                this.g1 = false;
            }
            this.d9 = false;
            if (!c) {
                this.o1();
            }
            this.r7();
            if (this.g3) {
                this.g2 = false;
                this.g1 = true;
            }
            if (this.g6) {
                this.g5 = false;
                this.g4 = true;
            }
        }
        else {
            this.aah(this.yu, this.y3);
            this.aak(this.yu, this.y3);
        }
    };
    CategorySeries.prototype.c4 = function () {
        return 3;
    };
    CategorySeries.prototype.c7 = function () {
        var a = this.c4();
        switch (a) {
            case 0: return 0;
            case 1: return 1;
            case 4: return 4;
            case 2: return 2;
            case 3: return 3;
        }
        return 3;
    };
    CategorySeries.prototype.zu = function () {
        if (this.isVertical) {
            return this.ws.left;
        }
        return this.ws.bottom;
    };
    CategorySeries.prototype.zo = function () {
        return false;
    };
    CategorySeries.prototype.ht = function () {
        var a = _super.prototype.ht.call(this);
        if (this.zo()) {
            var b = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "DashArray";
                $ret.p = true;
                $ret.x = ((function () {
                    var $ret = new DoubleCollection();
                    $ret.add(5);
                    $ret.add(5);
                    return $ret;
                })());
                $ret.q = true;
                return $ret;
            })());
            a.add(b);
        }
        if (!this.zo()) {
            var c = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Brush";
                $ret.p = true;
                $ret.y = "ActualBrush";
                $ret.q = true;
                return $ret;
            })());
            c.valueResolving = delegateCombine(c.valueResolving, runOn(this, this.r0));
            a.add(c);
        }
        if (!this.zo()) {
            var d = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Outline";
                $ret.p = true;
                $ret.y = "ActualOutline";
                $ret.q = true;
                return $ret;
            })());
            d.valueResolving = delegateCombine(d.valueResolving, runOn(this, this.r0));
            a.add(d);
        }
        var e = ((function () {
            var $ret = new SeriesLayerPropertyOverlay();
            $ret.p = true;
            $ret.ad = "IsTransitionInEnabled";
            $ret.x = true;
            $ret.q = true;
            return $ret;
        })());
        a.add(e);
        var f = ((function () {
            var $ret = new SeriesLayerPropertyOverlay();
            $ret.p = true;
            $ret.ad = "TransitionInMode";
            $ret.x = enumGetBox(CategoryTransitionInMode_$type, 20);
            $ret.q = true;
            return $ret;
        })());
        a.add(f);
        return a;
    };
    CategorySeries.prototype.kv = function () {
        if (this.d3()) {
            return this.bn.yy;
        }
        else {
            return this.bn.yu;
        }
    };
    CategorySeries.prototype.zy = function () {
        var e_1, _a;
        if (this.bn != null) {
            var a = this.bn;
            return a.zy();
        }
        var b = 0;
        var c = this.fetchXAxis();
        try {
            for (var _b = __values(fromEnum(this.db.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var d = _c.value;
                if (d == this) {
                    return b;
                }
                var e = typeCast(CategorySeries.$, d);
                if (e != null) {
                    var f = typeCast(CategoryAxisBase.$, e.fetchXAxis());
                    if (f == c && e.preferredCategoryMode(f) == 2) {
                        b++;
                    }
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        return -1;
    };
    Object.defineProperty(CategorySeries.prototype, "GetErrorBarsXAxis", {
        get: function () {
            return this.fetchXAxis();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeries.prototype, "GetErrorBarsYAxis", {
        get: function () {
            return this.fetchYAxis();
        },
        enumerable: false,
        configurable: true
    });
    CategorySeries.prototype.z9 = function (a, b) {
        this.y2.a1(this, this.zn(), runOn(this, this.ym), this.zv(a), this.zw(a));
        var c = this.y2.d;
        var d = c != null;
        var e = this.ys();
        if (e == null) {
            return;
        }
        var f = this.getEffectiveViewport1(a);
        var g = new ScalerParams(0, a.by, a.bx, e.ch, f);
        var h = this.dn == null ? 0 : this.dn.count;
        var i = a;
        var j = b.f;
        var k = i.cw.e;
        var l = i.cw.f;
        var m = i.cw.d;
        var n = k * m;
        for (var o = 0; o < a.cz.count; o++) {
            var p = a.cz.item(o);
            var q = p.content;
            if (d) {
                var r = p.al;
                this.aaf(j, r, h, e, g, a.q, p.ak);
            }
            this.y2.ac(p, q);
        }
    };
    CategorySeries.prototype.aak = function (a, b) {
        this.qx();
        this.wi = a.ab;
        b.aw();
    };
    CategorySeries.prototype.aah = function (a, b) {
        a.v();
    };
    CategorySeries.prototype.zv = function (a) {
        return a.cw.d;
    };
    CategorySeries.prototype.zw = function (a) {
        return a.cw.e;
    };
    CategorySeries.prototype.zx = function (a) {
        return a.cw.f;
    };
    CategorySeries.prototype.z8 = function (a) {
        var b = a;
        b.cz.count = 0;
    };
    CategorySeries.prototype.rt = function (a, b) {
        _super.prototype.rt.call(this, a, b);
        if (!this.g0) {
            this.cv.a4(b);
            return;
        }
        var c = typeCast(CategorySeriesView.$, this.cu);
        c.cw.g(this.id);
        this.cv.a4(b);
        if (this.d5(this.cu)) {
            return;
        }
        this.aal();
        this.g0 = false;
    };
    CategorySeries.prototype.aal = function () {
        var a = typeCast(CategorySeriesView.$, this.cu);
        if (!this.gs) {
            this.aah(this.yx, a);
        }
        this.gs = false;
        this.aak(this.yx, a);
    };
    CategorySeries.prototype.y0 = function (a) {
        var b = typeCast(CategorySeriesView.$, a);
        if (b != null && b.q) {
            if (b.p) {
                return new CategoryFramePreparer(1, typeCast(IIsCategoryBased_$type, this), typeCast(ISupportsMarkers_$type, b), b, this, b.cw);
            }
            else {
                return new CategoryFramePreparer(1, typeCast(IIsCategoryBased_$type, this), typeCast(ISupportsMarkers_$type, b), this.db.c9.o, this, b.cw);
            }
        }
        else {
            return this.yz;
        }
    };
    CategorySeries.prototype.aw = function (a) {
        if (this.fetchXAxis() == a) {
            return this.fetchYAxis();
        }
        if (this.fetchYAxis() == a) {
            return this.fetchXAxis();
        }
        return null;
    };
    CategorySeries.prototype.ai = function () {
        return [this.fetchXAxis(), this.fetchYAxis()];
    };
    CategorySeries.yn = function (a, b, c, d, e) {
        var f = null;
        if (b > -1 && !d.isEmpty && !c.isEmpty) {
            var g = a._inner[0][0] < c.left - 2000 ? c.left - 10 : NaN;
            var h = c.bottom + 10;
            var i = a._inner[b][0] > c.right + 2000 ? c.right + 10 : NaN;
            var j = c.top - 10;
            f = new Clipper(1, g, h, i, j, e);
        }
        return f;
    };
    CategorySeries.prototype.aac = function (a, b, c, d, e) {
        var f = this.wr(a);
        var g = this.ys();
        var h = this.y9();
        var i = this.fetchXAxis();
        var j = g == i;
        var k = this.wj(this.cv);
        var l, m;
        var n = this.aab(f, g, j, k, l, m);
        l = n.p4;
        m = n.p5;
        if (d == 0) {
            this.aae(b, l, m, c);
        }
        else {
            var o = void 0, p = void 0;
            var q = this.aaa(f, h, j, k, o, p);
            o = q.p4;
            p = q.p5;
            this.aad(e, b, l, m, o, p, c);
        }
    };
    CategorySeries.prototype.aae = function (a, b, c, d) {
        var _this = this;
        if (b == -1 || c == -1) {
            return;
        }
        if (d == 7 || d == 6) {
            a.add(((function () {
                var $ret = new ChartSelection();
                $ret.series = _this;
                return $ret;
            })()));
            return;
        }
        for (var e = b; e <= c; e++) {
            this.z7(a, e, d);
        }
    };
    CategorySeries.prototype.aaa = function (a, b, c, d, e, f) {
        var g = new ScalerParams(0, a, this.db.wk, b.ch, this.getEffectiveViewport());
        var h = c ? (b.ch ? d.top : d.bottom) : (b.ch ? d.right : d.left);
        var i = c ? (b.ch ? d.bottom : d.top) : (b.ch ? d.left : d.right);
        e = b.eo(h, g);
        f = b.eo(i, g);
        return {
            p4: e,
            p5: f
        };
    };
    CategorySeries.prototype.aab = function (a, b, c, d, e, f) {
        var g = new ScalerParams(0, a, this.db.wk, b.ch, this.getEffectiveViewport());
        var h = c ? (b.ch ? d.right : d.left) : (b.ch ? d.top : d.bottom);
        var i = c ? (b.ch ? d.left : d.right) : (b.ch ? d.bottom : d.top);
        var j = b.eo(h, g);
        var k = b.eo(i, g);
        e = Math.max(0, truncate((b.categoryMode == 0 ? Math.ceil(j) : Math.floor(j))));
        f = Math.min(truncate(Math.floor(k)), this.dn.count - 1);
        return {
            p4: e,
            p5: f
        };
    };
    CategorySeries.$t = markType(CategorySeries, 'CategorySeries', MarkerSeries.$, [IHasCategoryModePreference_$type, ISupportsErrorBars_$type]);
    CategorySeries.$$p = markDep(DependencyProperty, PropertyMetadata, CategorySeries, 'raisePropertyChanged', ['CategoryCollisionMode:yt:aam', [CategoryCollisionMode_$type, enumGetBox(CategoryCollisionMode_$type, 0)], 'IsCustomCategoryMarkerStyleAllowed:zj:aan', [0, false], 'IsCustomCategoryStyleAllowed:zk:aao', [0, false], 'IsTransitionInEnabled:zl:aap', [0, false], 'TransitionInMode:y5:aaq', [CategoryTransitionInMode_$type, enumGetBox(CategoryTransitionInMode_$type, 0)], 'UseHighMarkerFidelity:zq:aar', [0, false]]);
    return CategorySeries;
}(MarkerSeries));
export { CategorySeries };
/**
 * @hidden
 */
var CategorySeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategorySeriesView, _super);
    function CategorySeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.cy = null;
        _this.c1 = 0;
        _this.cw = null;
        _this.cz = null;
        _this.c1 = -1;
        _this.cy = a;
        _this.cw = _this.cx();
        _this.cz = ((function () {
            var $ret = new Pool$1(Marker.$);
            $ret.create = runOn(_this, _this.b3);
            $ret.activate = runOn(_this, _this.co);
            $ret.disactivate = runOn(_this, _this.cq);
            $ret.destroy = runOn(_this, _this.cp);
            return $ret;
        })());
        return _this;
    }
    CategorySeriesView.prototype.c0 = function (a) {
        if (this.c1 != a.u) {
            return true;
        }
        return false;
    };
    CategorySeriesView.prototype.c3 = function (a) {
        this.c1 = a.u;
    };
    CategorySeriesView.prototype.updateMarkerTemplate = function (a, b, c) {
        if (!this.b4.xi) {
            var d = this.cz.item(a);
            var e = d.content;
            e.item = this.e.dn.item(b);
            d.ak = b;
            d.al = c;
        }
    };
    CategorySeriesView.prototype.cx = function () {
        return new CategoryBucketCalculator(this);
    };
    Object.defineProperty(CategorySeriesView.prototype, "shouldDisplayMarkers", {
        get: function () {
            return this.cy.xg();
        },
        enumerable: false,
        configurable: true
    });
    CategorySeriesView.prototype.updateMarkerCount = function (a) {
        this.cz.count = a;
    };
    CategorySeriesView.prototype.cj = function (a) {
        this.cz.c(a);
    };
    CategorySeriesView.prototype.c2 = function () {
    };
    CategorySeriesView.$t = markType(CategorySeriesView, 'CategorySeriesView', MarkerSeriesView.$, [ISupportsMarkers_$type]);
    return CategorySeriesView;
}(MarkerSeriesView));
export { CategorySeriesView };
