// igniteui-react-charts/esm5/lib/ChaikinVolatilityIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ChaikinVolatilityIndicatorStrategy } from "./ChaikinVolatilityIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var ChaikinVolatilityIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ChaikinVolatilityIndicator, _super);
    function ChaikinVolatilityIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChaikinVolatilityIndicator.prototype.bj = function () {
        return new ChaikinVolatilityIndicator();
    };
    ChaikinVolatilityIndicator.prototype.get_abv = function () {
        return new ChaikinVolatilityIndicatorStrategy();
    };
    Object.defineProperty(ChaikinVolatilityIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    ChaikinVolatilityIndicator.prototype.get_ab8 = function () {
        return ChaikinVolatilityIndicator.$;
    };
    Object.defineProperty(ChaikinVolatilityIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ChaikinVolatilityIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(ChaikinVolatilityIndicator.ach));
        },
        set: function (a) {
            this.h(ChaikinVolatilityIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    ChaikinVolatilityIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    ChaikinVolatilityIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    ChaikinVolatilityIndicator.$t = markType(ChaikinVolatilityIndicator, 'ChaikinVolatilityIndicator', StrategyBasedIndicator.$);
    ChaikinVolatilityIndicator.ach = StrategyBasedIndicator.acd(10, ChaikinVolatilityIndicator.$);
    return ChaikinVolatilityIndicator;
}(StrategyBasedIndicator));
export { ChaikinVolatilityIndicator };
