// igniteui-react-charts/esm5/lib/ChaikinOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { ChaikinOscillatorIndicatorStrategy } from "./ChaikinOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var ChaikinOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ChaikinOscillatorIndicator, _super);
    function ChaikinOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChaikinOscillatorIndicator.prototype.bj = function () {
        return new ChaikinOscillatorIndicator();
    };
    ChaikinOscillatorIndicator.prototype.get_abv = function () {
        return new ChaikinOscillatorIndicatorStrategy();
    };
    Object.defineProperty(ChaikinOscillatorIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    ChaikinOscillatorIndicator.prototype.get_ab8 = function () {
        return ChaikinOscillatorIndicator.$;
    };
    Object.defineProperty(ChaikinOscillatorIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ChaikinOscillatorIndicator.prototype, "shortPeriod", {
        get: function () {
            return typeGetValue(this.c(ChaikinOscillatorIndicator.ach));
        },
        set: function (a) {
            this.h(ChaikinOscillatorIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    ChaikinOscillatorIndicator.prototype.ab3 = function () {
        return this.shortPeriod;
    };
    Object.defineProperty(ChaikinOscillatorIndicator.prototype, "longPeriod", {
        get: function () {
            return typeGetValue(this.c(ChaikinOscillatorIndicator.acg));
        },
        set: function (a) {
            this.h(ChaikinOscillatorIndicator.acg, a);
        },
        enumerable: false,
        configurable: true
    });
    ChaikinOscillatorIndicator.prototype.ab0 = function () {
        return this.longPeriod;
    };
    ChaikinOscillatorIndicator.$t = markType(ChaikinOscillatorIndicator, 'ChaikinOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    ChaikinOscillatorIndicator.ach = StrategyBasedIndicator.acf(3, ChaikinOscillatorIndicator.$);
    ChaikinOscillatorIndicator.acg = StrategyBasedIndicator.acb(10, ChaikinOscillatorIndicator.$);
    return ChaikinOscillatorIndicator;
}(StrategyBasedIndicator));
export { ChaikinOscillatorIndicator };
