// igniteui-react-charts/esm5/lib/CollisionAvoidanceType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available methods of collision avoidance of series' markers.
 */
export var CollisionAvoidanceType = /*@__PURE__*/ (function (CollisionAvoidanceType) {
    /**
     * Collision avoidance is disabled.
     */
    CollisionAvoidanceType[CollisionAvoidanceType["None"] = 0] = "None";
    /**
     * Items colliding with other items will be hidden from view.
     */
    CollisionAvoidanceType[CollisionAvoidanceType["Omit"] = 1] = "Omit";
    /**
     * Items colliding with other items will be partially hidden from view by reducing their opacity.
     */
    CollisionAvoidanceType[CollisionAvoidanceType["Fade"] = 2] = "Fade";
    /**
     * Items colliding with other items will be either hidden from view or moved to new positions.
     */
    CollisionAvoidanceType[CollisionAvoidanceType["OmitAndShift"] = 3] = "OmitAndShift";
    /**
     * Items colliding with other items will be either partially hidden from view by reducing their opacity, or moved to new positions, or a combination of both.
     */
    CollisionAvoidanceType[CollisionAvoidanceType["FadeAndShift"] = 4] = "FadeAndShift";
    return CollisionAvoidanceType;
})({});
/**
 * @hidden
 */
export var CollisionAvoidanceType_$type = /*@__PURE__*/ markEnum('CollisionAvoidanceType', 'None,0|Omit,1|Fade,2|OmitAndShift,3|FadeAndShift,4');
