// igniteui-react-charts/esm5/lib/CommodityChannelIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { CommodityChannelIndexIndicatorStrategy } from "./CommodityChannelIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var CommodityChannelIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CommodityChannelIndexIndicator, _super);
    function CommodityChannelIndexIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CommodityChannelIndexIndicator.prototype.bj = function () {
        return new CommodityChannelIndexIndicator();
    };
    CommodityChannelIndexIndicator.prototype.get_abv = function () {
        return new CommodityChannelIndexIndicatorStrategy();
    };
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    CommodityChannelIndexIndicator.prototype.get_ab8 = function () {
        return CommodityChannelIndexIndicator.$;
    };
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(CommodityChannelIndexIndicator.ach));
        },
        set: function (a) {
            this.h(CommodityChannelIndexIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    CommodityChannelIndexIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    CommodityChannelIndexIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    CommodityChannelIndexIndicator.$t = markType(CommodityChannelIndexIndicator, 'CommodityChannelIndexIndicator', StrategyBasedIndicator.$);
    CommodityChannelIndexIndicator.ach = StrategyBasedIndicator.acd(20, CommodityChannelIndexIndicator.$);
    return CommodityChannelIndexIndicator;
}(StrategyBasedIndicator));
export { CommodityChannelIndexIndicator };
