// igniteui-react-charts/esm5/lib/ConsolidatedItemsPosition.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Enum used to determine positioning logic for data items which have been consolidated into a single visual element.
 */
export var ConsolidatedItemsPosition = /*@__PURE__*/ (function (ConsolidatedItemsPosition) {
    /**
     * Consolidated Items should be positioned using their minimum value.
     */
    ConsolidatedItemsPosition[ConsolidatedItemsPosition["Minimum"] = 0] = "Minimum";
    /**
     * Consolidated Items should be positioned using their maximum value.
     */
    ConsolidatedItemsPosition[ConsolidatedItemsPosition["Maximum"] = 1] = "Maximum";
    /**
     * Consolidated Items should be positioned at the midpoint of the range.
     */
    ConsolidatedItemsPosition[ConsolidatedItemsPosition["Median"] = 2] = "Median";
    /**
     * Consolidated Items should be positioned using the value nearest to the reference value of the corresponding axis.
     */
    ConsolidatedItemsPosition[ConsolidatedItemsPosition["RelativeMinimum"] = 3] = "RelativeMinimum";
    /**
     * Consolidated Items should be positioned using the value farthest from the reference value of the corresponding axis.
     */
    ConsolidatedItemsPosition[ConsolidatedItemsPosition["RelativeMaximum"] = 4] = "RelativeMaximum";
    return ConsolidatedItemsPosition;
})({});
/**
 * @hidden
 */
export var ConsolidatedItemsPosition_$type = /*@__PURE__*/ markEnum('ConsolidatedItemsPosition', 'Minimum,0|Maximum,1|Median,2|RelativeMinimum,3|RelativeMaximum,4');
