// igniteui-react-charts/esm5/lib/DataTooltipConstraintMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available options for constraining the tooltip to different bounds.
 */
export var DataTooltipConstraintMode = /*@__PURE__*/ (function (DataTooltipConstraintMode) {
    /**
     * A constraint for the tooltip is selected automatically.
     */
    DataTooltipConstraintMode[DataTooltipConstraintMode["Auto"] = 0] = "Auto";
    /**
     * The tooltip is constrained within the bounds of the application window.
     */
    DataTooltipConstraintMode[DataTooltipConstraintMode["Application"] = 1] = "Application";
    /**
     * The tooltip is constrained within the bounds of the chart.
     */
    DataTooltipConstraintMode[DataTooltipConstraintMode["Chart"] = 2] = "Chart";
    /**
     * The tooltip is constrained within the bounds of the chart plot area.
     */
    DataTooltipConstraintMode[DataTooltipConstraintMode["PlotArea"] = 3] = "PlotArea";
    /**
     * The tooltip is constrained within the bounds of the screen.
     */
    DataTooltipConstraintMode[DataTooltipConstraintMode["None"] = 4] = "None";
    return DataTooltipConstraintMode;
})({});
/**
 * @hidden
 */
export var DataTooltipConstraintMode_$type = /*@__PURE__*/ markEnum('DataTooltipConstraintMode', 'Auto,0|Application,1|Chart,2|PlotArea,3|None,4');
