// igniteui-react-charts/esm5/lib/DataToolTipLayer_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { AnnotationLayer } from "./AnnotationLayer";
import { Axis } from "./Axis";
import { DependencyProperty } from "igniteui-react-core";
import { DataToolTipLayerGroupingMode_$type } from "igniteui-react-core";
import { Series } from "./Series";
import { Rect } from "igniteui-react-core";
import { Base, enumGetBox, EnumUtil, fromEnum, typeCast, String_$type, typeGetValue, Point_$type, Array_$type, markType, markDep, runOn, TypeRegistrar } from "igniteui-react-core";
import { Size } from "igniteui-react-core";
import { RectUtil } from "igniteui-react-core";
import { DataTooltipGroupedPositionX_$type } from "igniteui-react-core";
import { DataTooltipGroupedPositionY_$type } from "igniteui-react-core";
import { DataAbbreviationMode_$type } from "igniteui-react-core";
import { DataLegendValueMode_$type } from "igniteui-react-core";
import { Thickness } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
import { DataLegendHeaderDateMode_$type } from "igniteui-react-core";
import { DataLegendHeaderTimeMode_$type } from "igniteui-react-core";
import { DataLegendSummaryType_$type } from "igniteui-react-core";
import { LegendItemBadgeShape_$type } from "igniteui-react-core";
import { DataLegendLayoutMode_$type } from "igniteui-react-core";
import { DataLegendUnitsMode_$type } from "igniteui-react-core";
import { DataLegendLabelMode_$type } from "igniteui-react-core";
import { DataToolTipLayerFrame } from "./DataToolTipLayerFrame";
import { Dictionary$2 } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { Pool$1 } from "igniteui-react-core";
import { DataToolTipSeriesInfo } from "./DataToolTipSeriesInfo";
import { DataContext } from "igniteui-react-core";
import { DataTooltipCollisionInfo } from "./DataTooltipCollisionInfo";
import { DeviceUtils } from "igniteui-react-core";
import { HashSet$1 } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
import { stringEmpty } from "igniteui-react-core";
import { AnnotationLayerView } from "./AnnotationLayerView";
import { PointerTooltip } from "./PointerTooltip";
import { IOnDemandRender_$type } from "igniteui-react-core";
/**
 * @hidden
 */
var DataToolTipLayer = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataToolTipLayer, _super);
    function DataToolTipLayer() {
        var _this = _super.call(this) || this;
        _this.aan = 1;
        _this.abs = new Dictionary$2(String_$type, Base.$, 0);
        _this.abt = new List$1(String_$type, 0);
        _this.aaf = 8;
        _this.aaj = 11;
        _this.aa3 = null;
        _this.aav = null;
        _this.aau = null;
        _this.aay = null;
        _this.aa0 = null;
        _this.aax = null;
        _this.aaz = null;
        _this.aa2 = null;
        _this.aa1 = null;
        _this.aaw = null;
        var a = new DataToolTipLayerFrame();
        var b = new DataToolTipLayerFrame();
        var c = new DataToolTipLayerFrame();
        _this.xc = a;
        _this.xb = b;
        _this.xd = c;
        _this.ab = DataToolTipLayer.$;
        return _this;
    }
    DataToolTipLayer.prototype.bj = function () {
        return new DataToolTipLayer();
    };
    DataToolTipLayer.prototype.get_er = function () {
        return true;
    };
    DataToolTipLayer.prototype.cs = function () {
        return new DataToolTipLayerView(this);
    };
    DataToolTipLayer.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.aar = a;
    };
    DataToolTipLayer.prototype.get_e2 = function () {
        return true;
    };
    Object.defineProperty(DataToolTipLayer.prototype, "aao", {
        get: function () {
            return this.aan;
        },
        set: function (a) {
            var b = this.aan;
            this.aan = a;
            if (b != this.aan) {
                this.raisePropertyChanged("ActualGroupingMode", enumGetBox(DataToolTipLayerGroupingMode_$type, b), enumGetBox(DataToolTipLayerGroupingMode_$type, this.aan));
            }
        },
        enumerable: false,
        configurable: true
    });
    DataToolTipLayer.prototype.aaq = function () {
        var e_1, _a;
        var a = true;
        var b = true;
        var c = true;
        var d = true;
        if (this.aap != 0) {
            return this.aap;
        }
        var e = this.z5();
        if (e != null) {
            try {
                for (var _b = __values(fromEnum(e.db())), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var f = _c.value;
                    if (!f.es && !f.fw) {
                        a = false;
                    }
                    if (!f.es && !f.fp) {
                        b = false;
                    }
                    if (!f.es && !f.fx) {
                        c = false;
                    }
                    if (!f.es && !f.ft) {
                        d = false;
                    }
                }
            }
            catch (e_1_1) {
                e_1 = { error: e_1_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_1)
                        throw e_1.error;
                }
            }
        }
        if (a || b || c || d) {
            return 2;
        }
        return 1;
    };
    DataToolTipLayer.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case DataToolTipLayer.$$p[42]:
                this.rr(true);
                break;
            case DataToolTipLayer.$$p[49]:
                this.rr(true);
                break;
            case DataToolTipLayer.$$p[12]:
                this.aao = this.aaq();
                break;
            case "ActualGroupingMode":
                this.rr(true);
                break;
            case DataToolTipLayer.$$p[10]:
                this.aag = this.aai();
                break;
            case DataToolTipLayer.$$p[11]:
                this.aak = this.aam();
                break;
            case "SeriesViewer":
                this.aar.cl();
                break;
            case DataToolTipLayer.$$p[31]:
            case DataToolTipLayer.$$p[24]:
            case DataToolTipLayer.$$p[5]:
            case DataToolTipLayer.$$p[23]:
            case DataToolTipLayer.$$p[4]:
            case DataToolTipLayer.$$p[28]:
            case DataToolTipLayer.$$p[50]:
            case DataToolTipLayer.$$p[52]:
            case DataToolTipLayer.$$p[53]:
            case DataToolTipLayer.$$p[54]:
            case DataToolTipLayer.$$p[51]:
            case DataToolTipLayer.$$p[57]:
            case DataToolTipLayer.$$p[56]:
            case DataToolTipLayer.$$p[55]:
            case DataToolTipLayer.$$p[58]:
            case DataToolTipLayer.$$p[59]:
            case DataToolTipLayer.$$p[63]:
            case DataToolTipLayer.$$p[62]:
            case DataToolTipLayer.$$p[61]:
            case DataToolTipLayer.$$p[60]:
            case "ValueTextStyle":
            case DataToolTipLayer.$$p[16]:
            case DataToolTipLayer.$$p[15]:
            case DataToolTipLayer.$$p[13]:
            case DataToolTipLayer.$$p[14]:
            case DataToolTipLayer.$$p[17]:
            case DataToolTipLayer.$$p[22]:
            case DataToolTipLayer.$$p[20]:
            case DataToolTipLayer.$$p[21]:
            case DataToolTipLayer.$$p[18]:
            case DataToolTipLayer.$$p[19]:
            case "HeaderTextStyle":
            case DataToolTipLayer.$$p[8]:
            case DataToolTipLayer.$$p[9]:
            case DataToolTipLayer.$$p[6]:
            case DataToolTipLayer.$$p[7]:
            case "GroupTextStyle":
            case DataToolTipLayer.$$p[35]:
            case "SummaryTitleTextStyle":
            case DataToolTipLayer.$$p[38]:
            case DataToolTipLayer.$$p[37]:
            case DataToolTipLayer.$$p[36]:
            case DataToolTipLayer.$$p[34]:
            case DataToolTipLayer.$$p[41]:
            case "SummaryValueTextStyle":
            case DataToolTipLayer.$$p[33]:
            case DataToolTipLayer.$$p[32]:
            case "SummaryLabelTextStyle":
            case DataToolTipLayer.$$p[40]:
            case DataToolTipLayer.$$p[39]:
            case "SummaryUnitsTextStyle":
            case DataToolTipLayer.$$p[0]:
            case DataToolTipLayer.$$p[1]:
            case DataToolTipLayer.$$p[45]:
            case DataToolTipLayer.$$p[48]:
            case DataToolTipLayer.$$p[46]:
            case DataToolTipLayer.$$p[47]:
            case "UnitsTextStyle":
            case DataToolTipLayer.$$p[44]:
            case DataToolTipLayer.$$p[43]:
            case "TitleTextStyle":
            case DataToolTipLayer.$$p[25]:
            case DataToolTipLayer.$$p[26]:
            case DataToolTipLayer.$$p[27]:
            case "LabelTextStyle":
                this.afg(b, c, d);
                break;
        }
    };
    DataToolTipLayer.prototype.afg = function (a, b, c) {
        if (!this.abs.containsKey(a)) {
            this.abt.add(a);
            this.abs.addItem(a, c);
        }
        else {
            this.abs.item(a, c);
        }
        this.rr(true);
    };
    DataToolTipLayer.prototype.afk = function (a, b) {
        switch (a) {
            case DataToolTipLayer.$$p[12]:
                this.aap = EnumUtil.getEnumValue(DataToolTipLayerGroupingMode_$type, b);
                break;
            case DataToolTipLayer.$$p[29]:
                this.abx = b;
                break;
            case DataToolTipLayer.$$p[30]:
                this.aby = b;
                break;
            case DataToolTipLayer.$$p[2]:
                this.abv = b;
                break;
            case DataToolTipLayer.$$p[3]:
                this.abw = b;
                break;
            case DataToolTipLayer.$$p[10]:
                this.aah = EnumUtil.getEnumValue(DataTooltipGroupedPositionX_$type, b);
                break;
            case DataToolTipLayer.$$p[11]:
                this.aal = EnumUtil.getEnumValue(DataTooltipGroupedPositionY_$type, b);
                break;
            case DataToolTipLayer.$$p[31]:
                this.abn = b;
                break;
            case DataToolTipLayer.$$p[24]:
                this.z3 = b;
                break;
            case DataToolTipLayer.$$p[5]:
                this.zy = b;
                break;
            case DataToolTipLayer.$$p[23]:
                this.z2 = b;
                break;
            case DataToolTipLayer.$$p[4]:
                this.zx = b;
                break;
            case DataToolTipLayer.$$p[50]:
                this.z7 = EnumUtil.getEnumValue(DataAbbreviationMode_$type, b);
                break;
            case DataToolTipLayer.$$p[52]:
                this.acx = typeGetValue(b);
                break;
            case DataToolTipLayer.$$p[53]:
                this.acy = typeGetValue(b);
                break;
            case DataToolTipLayer.$$p[54]:
                this.aae = EnumUtil.getEnumValue(DataLegendValueMode_$type, b);
                break;
            case DataToolTipLayer.$$p[51]:
                this.aey = b;
                break;
            case DataToolTipLayer.$$p[57]:
                this.abp = b;
                break;
            case DataToolTipLayer.$$p[56]:
                this.ae4 = b;
                break;
            case DataToolTipLayer.$$p[55]:
                this.z4 = b;
                break;
            case DataToolTipLayer.$$p[58]:
                this.ah3 = b;
                break;
            case DataToolTipLayer.$$p[59]:
                this.abq = b;
                break;
            case DataToolTipLayer.$$p[63]:
                this.afd = b;
                break;
            case DataToolTipLayer.$$p[62]:
                this.abr = b;
                break;
            case DataToolTipLayer.$$p[61]:
                this.ah4 = b;
                break;
            case DataToolTipLayer.$$p[60]:
                this.ahm = b;
                break;
            case "ValueTextStyle":
                this.abd = b;
                break;
            case DataToolTipLayer.$$p[16]:
                this.adl = b;
                break;
            case DataToolTipLayer.$$p[15]:
                this.z1 = b;
                break;
            case DataToolTipLayer.$$p[13]:
                this.adh = b;
                break;
            case DataToolTipLayer.$$p[14]:
                this.z8 = EnumUtil.getEnumValue(DataLegendHeaderDateMode_$type, b);
                break;
            case DataToolTipLayer.$$p[17]:
                this.z9 = EnumUtil.getEnumValue(DataLegendHeaderTimeMode_$type, b);
                break;
            case DataToolTipLayer.$$p[22]:
                this.adq = b;
                break;
            case DataToolTipLayer.$$p[20]:
                this.ahe = b;
                break;
            case DataToolTipLayer.$$p[21]:
                this.ahx = b;
                break;
            case DataToolTipLayer.$$p[18]:
                this.ahw = b;
                break;
            case DataToolTipLayer.$$p[19]:
                this.abg = b;
                break;
            case "HeaderTextStyle":
                this.aa5 = b;
                break;
            case DataToolTipLayer.$$p[28]:
                this.aab = EnumUtil.getEnumValue(DataLegendLayoutMode_$type, b);
                break;
            case DataToolTipLayer.$$p[8]:
                this.ahd = b;
                break;
            case DataToolTipLayer.$$p[9]:
                this.ahv = b;
                break;
            case DataToolTipLayer.$$p[6]:
                this.ahu = b;
                break;
            case DataToolTipLayer.$$p[7]:
                this.abf = b;
                break;
            case "GroupTextStyle":
                this.aa4 = b;
                break;
            case DataToolTipLayer.$$p[35]:
                this.ahh = b;
                break;
            case "SummaryTitleTextStyle":
                this.aa8 = b;
                break;
            case DataToolTipLayer.$$p[38]:
                this.aac = EnumUtil.getEnumValue(DataLegendSummaryType_$type, b);
                break;
            case DataToolTipLayer.$$p[37]:
                this.aea = b;
                break;
            case DataToolTipLayer.$$p[36]:
                this.ah0 = b;
                break;
            case DataToolTipLayer.$$p[34]:
                this.ahz = b;
                break;
            case DataToolTipLayer.$$p[41]:
                this.ahj = b;
                break;
            case "SummaryValueTextStyle":
                this.aba = b;
                break;
            case DataToolTipLayer.$$p[33]:
                this.ad5 = b;
                break;
            case DataToolTipLayer.$$p[32]:
                this.ahg = b;
                break;
            case "SummaryLabelTextStyle":
                this.aa7 = b;
                break;
            case DataToolTipLayer.$$p[40]:
                this.aeg = b;
                break;
            case DataToolTipLayer.$$p[39]:
                this.ahi = b;
                break;
            case "SummaryUnitsTextStyle":
                this.aa9 = b;
                break;
            case DataToolTipLayer.$$p[0]:
                this.aht = b;
                break;
            case DataToolTipLayer.$$p[1]:
                this.aat = EnumUtil.getEnumValue(LegendItemBadgeShape_$type, b);
                break;
            case DataToolTipLayer.$$p[45]:
                this.aad = EnumUtil.getEnumValue(DataLegendUnitsMode_$type, b);
                break;
            case DataToolTipLayer.$$p[48]:
                this.aer = b;
                break;
            case DataToolTipLayer.$$p[46]:
                this.ahl = b;
                break;
            case DataToolTipLayer.$$p[47]:
                this.ah2 = b;
                break;
            case "UnitsTextStyle":
                this.abc = b;
                break;
            case DataToolTipLayer.$$p[44]:
                this.ah1 = b;
                break;
            case DataToolTipLayer.$$p[43]:
                this.ahk = b;
                break;
            case "TitleTextStyle":
                this.abb = b;
                break;
            case DataToolTipLayer.$$p[25]:
                this.aaa = EnumUtil.getEnumValue(DataLegendLabelMode_$type, b);
                break;
            case DataToolTipLayer.$$p[26]:
                this.ahf = b;
                break;
            case DataToolTipLayer.$$p[27]:
                this.ahy = b;
                break;
            case "LabelTextStyle":
                this.aa6 = b;
                break;
        }
    };
    DataToolTipLayer.prototype.afh = function (a) {
        for (var b = 0; b < this.abt.count; b++) {
            var c = this.abt._inner[b];
            a.updatePropertyByName(c, this.abs.item(c));
        }
    };
    DataToolTipLayer.prototype.yo = function (a, b) {
        {
            if (!this.a1.c) {
                this.rr(b);
            }
        }
    };
    DataToolTipLayer.prototype.yp = function (a, b) {
        if (!this.a1.c) {
            this.rr(b);
        }
    };
    DataToolTipLayer.prototype.yr = function (a, b) {
        var e_2, _a;
        var _this = this;
        _super.prototype.yr.call(this, a, b);
        this.aan = this.aaq();
        var c = a;
        var d = this.aar.b9;
        var e = this.abo;
        var f = false;
        c.m.clear();
        c.i.clear();
        c.l.clear();
        c.g.clear();
        c.f.clear();
        c.k.clear();
        c.j.clear();
        c.h.clear();
        if (isNaN_(this.zu.x) && isNaN_(this.zu.y)) {
            d.count = 0;
            return;
        }
        {
            if (this.aao == 2) {
                var g_1 = 0;
                {
                    try {
                        for (var _b = __values(fromEnum(this.db.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                            var h = _c.value;
                            if (this.xi(h)) {
                                this.afj(h, c, e, f, g_1);
                                g_1++;
                            }
                            if (h.isStacked) {
                                h.p3(function (i) {
                                    if (_this.xi(i)) {
                                        _this.afj(i, c, e, f, g_1);
                                        g_1++;
                                    }
                                    if (i.eg) {
                                        i.p1(function (j) {
                                            if (_this.xi(j)) {
                                                _this.afj(j, c, e, f, g_1);
                                                g_1++;
                                            }
                                            return true;
                                        });
                                    }
                                    return true;
                                });
                            }
                            if (h.eg) {
                                h.p1(function (i) {
                                    if (_this.xi(i)) {
                                        _this.afj(i, c, e, f, g_1);
                                        g_1++;
                                    }
                                    return true;
                                });
                            }
                        }
                    }
                    catch (e_2_1) {
                        e_2 = { error: e_2_1 };
                    }
                    finally {
                        try {
                            if (_c && !_c.done && (_a = _b.return))
                                _a.call(_b);
                        }
                        finally {
                            if (e_2)
                                throw e_2.error;
                        }
                    }
                    this.aff(c);
                }
                d.count = g_1;
            }
            else {
                var i = this.z5();
                var j = this.abu(i);
                var k = this.aas(i, j, e);
                this.afi(k, c, e);
            }
        }
    };
    DataToolTipLayer.prototype.abu = function (a) {
        var e_3, _a;
        var b = new List$1(Series.$, 0);
        if (a == null) {
            return b;
        }
        try {
            for (var _b = __values(fromEnum(a.df)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                b.add(c);
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
        return b;
    };
    DataToolTipLayer.prototype.z5 = function () {
        var e_4, _a, e_5, _b;
        if (this.z6 != null) {
            return this.z6;
        }
        if (this.db == null) {
            return null;
        }
        try {
            for (var _c = __values(fromEnum(this.db.series)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var a = _d.value;
                if (a.ey) {
                    var b = a.a0();
                    if (b != null && b.co()) {
                        return b;
                    }
                }
            }
        }
        catch (e_4_1) {
            e_4 = { error: e_4_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_4)
                    throw e_4.error;
            }
        }
        try {
            for (var _e = __values(fromEnum(this.db.series)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var c = _f.value;
                if (!c.ey) {
                    var d = c.ai();
                    var e = d != null && d.length > 0 ? d[0] : null;
                    if (e != null && e.co()) {
                        return e;
                    }
                }
            }
        }
        catch (e_5_1) {
            e_5 = { error: e_5_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_5)
                    throw e_5.error;
            }
        }
        return null;
    };
    DataToolTipLayer.prototype.aas = function (a, b, c) {
        var d = 1.7976931348623157E+308;
        var e = -1.7976931348623157E+308;
        var f = new List$1(DataContext.$, 0);
        var g = new List$1(Base.$, 0);
        var h = false;
        var i = true;
        var j = { $type: Point_$type, x: NaN, y: NaN };
        var k = 0;
        var l = 0;
        var _loop_1 = function (m) {
            var n = b.item(m);
            k = n.db.l;
            l = n.db.m;
            if (n.ey) {
                var o = n;
                var p_1 = new List$1(Series.$, 0);
                if (o.isStacked) {
                    o.p3(function (q) {
                        p_1.add(q);
                        if (q.eg) {
                            q.p1(function (r) {
                                p_1.add(r);
                                return true;
                            });
                        }
                        return true;
                    });
                }
                else if (o.eg) {
                    p_1.add(o);
                    o.p1(function (q) {
                        p_1.add(q);
                        return true;
                    });
                }
                else {
                    p_1.add(o);
                }
                for (var q = 0; q < p_1.count; q++) {
                    var r = p_1._inner[q];
                    var s = p_1._inner[q];
                    if (s.db == null) {
                        continue;
                    }
                    if (!r.categoryAxis.co()) {
                        continue;
                    }
                    if (i) {
                        h = s.isVertical;
                    }
                    else {
                        if (h != s.isVertical) {
                            continue;
                        }
                    }
                    var t = this_1.yk(s);
                    if (!t.c) {
                        continue;
                    }
                    j = s.v7(t.d);
                    var u = s.wc(t.d, c, false);
                    if (h) {
                        if (!isNaN_(u.y)) {
                            d = Math.min(d, u.y);
                            e = Math.max(e, u.y);
                        }
                    }
                    else {
                        if (!isNaN_(u.x)) {
                            d = Math.min(d, u.x);
                            e = Math.max(e, u.x);
                        }
                    }
                }
            }
            else {
                var v = n.ai();
                for (var w = 0; w < v.length; w++) {
                    if (!v[w].co()) {
                        continue;
                    }
                }
                var x = n.wc(this_1.zu, c, false);
                j = n.v7(this_1.zu);
                if (h) {
                    if (!isNaN_(x.y)) {
                        d = Math.min(d, x.y);
                        e = Math.max(e, x.y);
                    }
                }
                else {
                    if (!isNaN_(x.x)) {
                        d = Math.min(d, x.x);
                        e = Math.max(e, x.x);
                    }
                }
            }
            var y = null;
            var z = n.kn(this_1.zu);
            if (n.e3 && z != null) {
                y = n.a3(z);
            }
            else {
                y = new DataContext();
                y.item = z;
                y.series = n;
            }
            f.add(y);
        };
        var this_1 = this;
        for (var m = 0; m < b.count; m++) {
            _loop_1(m);
        }
        var aa = 0;
        var ab = 0;
        if (h) {
            ab = (d + e) / 2;
            aa = this.ws.right;
            if (!isNaN_(j.x)) {
                aa = j.x;
            }
            if (ab < 0 || ab > this.ws.bottom) {
                ab = NaN;
            }
        }
        else {
            aa = (d + e) / 2;
            if (!isNaN_(j.y)) {
                ab = j.y;
            }
            if (aa < 0 || aa > this.ws.right) {
                aa = NaN;
            }
        }
        var ac = new DataToolTipSeriesInfo();
        ac.f = { $type: Point_$type, x: aa, y: ab };
        ac.e = { $type: Point_$type, x: j.x, y: j.y };
        ac.b = f;
        ac.a = h;
        return ac;
    };
    DataToolTipLayer.prototype.p7 = function () {
        this.zu = { $type: Point_$type, x: NaN, y: NaN };
        this.rr(false);
        this.aar.cj(false);
    };
    DataToolTipLayer.prototype.p8 = function () {
        this.zu = { $type: Point_$type, x: NaN, y: NaN };
        this.rr(false);
        this.aar.cj(true);
    };
    Object.defineProperty(DataToolTipLayer.prototype, "ahq", {
        get: function () {
            var a = RectUtil.n(this.cv.bx);
            var b = new Size(1, this.cv.bx.width, this.cv.bx.height);
            return new Rect(3, a, b);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "ahp", {
        get: function () {
            var a = RectUtil.n(this.cv.bx);
            var b = new Size(1, this.cv.bx.width, this.cv.bx.height);
            return new Rect(3, this.aar.cm(a), b);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "aho", {
        get: function () {
            var a = new Rect(0, 0, 0, this.db.m, this.db.l);
            var b = RectUtil.n(a);
            var c = new Size(1, a.width, a.height);
            return new Rect(3, this.aar.cm(b), c);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "ahs", {
        get: function () {
            var a = { $type: Point_$type, x: 0, y: 0 };
            return new Rect(3, a, this.aar.cr());
        },
        enumerable: false,
        configurable: true
    });
    DataToolTipLayer.prototype.ahr = function () {
        return this.ahq;
    };
    DataToolTipLayer.prototype.abk = function (a, b, c) {
        if (Rect.l_op_Equality(a, Rect.empty)) {
            return false;
        }
        var d = a.top;
        return c < d;
    };
    DataToolTipLayer.prototype.abh = function (a, b, c, d) {
        if (Rect.l_op_Equality(a, Rect.empty)) {
            return false;
        }
        var e = a.bottom;
        return c + d.height >= e;
    };
    DataToolTipLayer.prototype.abi = function (a, b, c) {
        if (Rect.l_op_Equality(a, Rect.empty)) {
            return false;
        }
        var d = a.left;
        return c < d;
    };
    DataToolTipLayer.prototype.abj = function (a, b, c, d) {
        if (Rect.l_op_Equality(a, Rect.empty)) {
            return false;
        }
        var e = a.right;
        return c + d.width > e;
    };
    DataToolTipLayer.prototype.acn = function (a, b) {
        return a.top + b;
    };
    DataToolTipLayer.prototype.acl = function (a, b, c) {
        return a.bottom - c.height - b;
    };
    DataToolTipLayer.prototype.acm = function (a, b) {
        return ((a.bottom + a.top) / 2) - (b.height / 2);
    };
    DataToolTipLayer.prototype.acw = function (a, b, c, d, e) {
        var f = c.y - d.height - b;
        return f;
    };
    DataToolTipLayer.prototype.acu = function (a, b, c, d, e) {
        var f = c.y + b;
        return f;
    };
    DataToolTipLayer.prototype.acv = function (a, b, c, d, e) {
        var f = c.y - (d.height / 2);
        return f;
    };
    DataToolTipLayer.prototype.acq = function (a, b, c, d, e) {
        var f = c.y - d.height;
        f = f - b;
        return f;
    };
    DataToolTipLayer.prototype.aco = function (a, b, c, d, e) {
        var f = c.y;
        f = f + b;
        return f;
    };
    DataToolTipLayer.prototype.acp = function (a, b, c, d, e) {
        var f = c.y - (d.height / 2);
        return f;
    };
    DataToolTipLayer.prototype.ab6 = function (a, b) {
        return a.left + b;
    };
    DataToolTipLayer.prototype.ab8 = function (a, b, c) {
        return (a.right - (c.width)) - b;
    };
    DataToolTipLayer.prototype.ab7 = function (a, b, c) {
        return ((a.right + a.left) / 2) - (c.width / 2) + b;
    };
    DataToolTipLayer.prototype.acf = function (a, b, c, d, e) {
        var f = (c.x - d.width) - b;
        return f;
    };
    DataToolTipLayer.prototype.ach = function (a, b, c, d, e) {
        var f = c.x + b;
        return f;
    };
    DataToolTipLayer.prototype.acg = function (a, b, c, d, e) {
        var f = (c.x - (d.width / 2));
        f = f + b;
        return f;
    };
    DataToolTipLayer.prototype.acc = function (a, b, c, d, e) {
        var f = c.x - d.width;
        f = f - b;
        return f;
    };
    DataToolTipLayer.prototype.ace = function (a, b, c, d, e) {
        var f = c.x;
        f = f + b;
        return f;
    };
    DataToolTipLayer.prototype.acd = function (a, b, c, d, e) {
        var f = c.x - (d.width / 2);
        f = f + b;
        return f;
    };
    DataToolTipLayer.prototype.ab2 = function (a, b, c, d, e, f, g, h) {
        switch (a) {
            case 1:
                {
                    return this.acq(e, b, h, f, true);
                }
            case 2:
                {
                    return this.aco(e, b, h, f, true);
                }
            case 3:
                {
                    return this.acp(e, b, h, f, true);
                }
            case 5: return this.acr(e, c, b, h, f);
            case 6: return this.acs(e, c, b, h, f);
            case 4: return this.act(e, c, b, h, f);
            case 8: return this.aci(e, c, b, h, f);
            case 9: return this.acj(e, c, b, h, f);
            case 7: return this.ack(e, c, b, h, f);
            case 13: return this.acn(d, b);
            case 14: return this.acl(d, b, f);
            case 15: return this.acm(d, f);
            case 10: return this.acw(e, b, g, f, true);
            case 12: return this.acv(e, b, g, f, true);
            case 11:
            default: return this.acu(e, b, g, f, true);
        }
    };
    DataToolTipLayer.prototype.ab9 = function (a, b, c, d, e) {
        var f = d.x;
        f = f + b / 2;
        f = f - e.width;
        f = f - c;
        return f;
    };
    DataToolTipLayer.prototype.aca = function (a, b, c, d, e) {
        var f = d.x;
        f = f + b / 2;
        f = f - e.width / 2;
        f = f + c;
        return f;
    };
    DataToolTipLayer.prototype.acb = function (a, b, c, d, e) {
        var f = d.x;
        f = f + b / 2;
        f = f + c;
        return f;
    };
    DataToolTipLayer.prototype.ab3 = function (a, b, c, d, e) {
        var f = d.x;
        f = f - b / 2;
        f = f - e.width;
        f = f - c;
        return f;
    };
    DataToolTipLayer.prototype.ab4 = function (a, b, c, d, e) {
        var f = d.x;
        f = f - b / 2;
        f = f - e.width / 2;
        f = f + c;
        return f;
    };
    DataToolTipLayer.prototype.ab5 = function (a, b, c, d, e) {
        var f = d.x;
        f = f - b / 2;
        f = f + c;
        return f;
    };
    DataToolTipLayer.prototype.ack = function (a, b, c, d, e) {
        var f = d.y;
        f = f + b / 2;
        f = f - e.height;
        f = f - c;
        return f;
    };
    DataToolTipLayer.prototype.acj = function (a, b, c, d, e) {
        var f = d.y;
        f = f + b / 2;
        f = f - e.height / 2;
        f = f + c;
        return f;
    };
    DataToolTipLayer.prototype.aci = function (a, b, c, d, e) {
        var f = d.y;
        f = f + b / 2;
        f = f + c;
        return f;
    };
    DataToolTipLayer.prototype.act = function (a, b, c, d, e) {
        var f = d.y;
        f = f - b / 2;
        f = f - e.height;
        f = f - c;
        return f;
    };
    DataToolTipLayer.prototype.acs = function (a, b, c, d, e) {
        var f = d.y;
        f = f - b / 2;
        f = f - e.height / 2;
        f = f + c;
        return f;
    };
    DataToolTipLayer.prototype.acr = function (a, b, c, d, e) {
        var f = d.y;
        f = f - b / 2;
        f = f + c;
        return f;
    };
    DataToolTipLayer.prototype.ab1 = function (a, b, c, d, e, f, g, h) {
        switch (a) {
            case 10: return this.acf(e, b, g, f, true);
            case 11: return this.ach(e, b, g, f, true);
            case 12: return this.acg(e, b, g, f, true);
            case 13: return this.ab6(d, b);
            case 14: return this.ab8(e, b, f);
            case 15: return this.ab7(e, b, f);
            case 5: return this.ab5(e, c, b, h, f);
            case 6: return this.ab4(e, c, b, h, f);
            case 4: return this.ab3(e, c, b, h, f);
            case 8: return this.acb(e, c, b, h, f);
            case 9: return this.aca(e, c, b, h, f);
            case 7: return this.ab9(e, c, b, h, f);
            case 1:
                {
                    return this.acc(e, b, h, f, true);
                }
            case 3:
                {
                    return this.acd(e, b, h, f, true);
                }
            case 2:
            default:
                {
                    return this.ace(e, b, h, f, true);
                }
        }
    };
    DataToolTipLayer.prototype.ahn = function (a, b, c, d) {
        var e = this.aag;
        var f = this.aak;
        var g = this.z5();
        var h = 0;
        if (g != null && g.cb && this.db != null) {
            h = g.getCategorySize(this.db.v4, this.ws, this.getEffectiveViewport());
            if (g.categoryMode == 0) {
                h = 0;
            }
        }
        var i = this.abz(e);
        var j = this.ab0(f);
        var k = this.ab1(e, i, h, this.ahq, a, b, c, d);
        var l = this.ab2(f, j, h, this.ahq, a, b, c, d);
        var m = this.abl(a, i, k, b);
        var n = this.abm(a, j, l, b);
        if (!m) {
            var o = this.zz(e);
            var p = 0;
            while (!m && p < o.length) {
                e = o[p];
                i = this.abz(e);
                k = this.ab1(e, i, h, this.ahq, a, b, c, d);
                m = this.abl(a, i, k, b);
                p++;
            }
        }
        if (!n) {
            var q = this.z0(f);
            var r = 0;
            while (!n && r < q.length) {
                f = q[r];
                j = this.ab0(f);
                l = this.ab2(f, j, h, this.ahq, a, b, c, d);
                n = this.abm(a, j, l, b);
                r++;
            }
        }
        return { $type: Point_$type, x: k, y: l };
    };
    DataToolTipLayer.prototype.ab0 = function (a) {
        if (!isNaN_(this.aby)) {
            return this.aby;
        }
        switch (a) {
            case 1: return this.abw;
            case 2: return this.abw;
            case 3: return 0;
            case 4: return this.abw;
            case 5: return this.abw;
            case 6: return this.abw;
            case 7: return this.abw;
            case 8: return this.abw;
            case 9: return 0;
            case 10: return this.abw;
            case 11: return this.abw;
            case 12: return 0;
            case 13: return this.abw;
            case 14: return this.abw;
            case 15: return 0;
        }
        return this.abw;
    };
    DataToolTipLayer.prototype.abz = function (a) {
        if (!isNaN_(this.abx)) {
            return this.abx;
        }
        switch (a) {
            case 4: return this.abv;
            case 6: return 0;
            case 5: return this.abv;
            case 13: return this.abv;
            case 15: return 0;
            case 14: return this.abv;
            case 7: return this.abv;
            case 9: return 0;
            case 8: return this.abv;
            case 1: return this.abv;
            case 3: return 0;
            case 2: return this.abv;
            case 10: return this.abv;
            case 12: return 0;
            case 11: return this.abv;
        }
        return this.abv;
    };
    DataToolTipLayer.prototype.z0 = function (a) {
        var b = [a];
        switch (a) {
            case 1:
                b = [2, 3, 13, 14];
                break;
            case 2:
                b = [1, 3, 14, 13];
                break;
            case 3:
                b = [1, 2];
                break;
            case 4:
                b = [8, 7, 3, 13, 14];
                break;
            case 5:
                b = [7, 8, 3, 13, 14];
                break;
            case 6:
                b = [9, 8, 4, 3];
                break;
            case 7:
                b = [5, 4, 2, 14, 13];
                break;
            case 8:
                b = [4, 5, 3, 14, 13];
                break;
            case 9:
                b = [6, 4, 8, 3];
                break;
            case 10:
                b = [11, 12, 2, 1, 3, 13, 14];
                break;
            case 11:
                b = [10, 12, 1, 2, 3, 14, 13];
                break;
            case 12:
                b = [10, 11];
                break;
            case 13: break;
            case 14: break;
            case 15: break;
        }
        return b;
    };
    DataToolTipLayer.prototype.zz = function (a) {
        var b = [a];
        switch (a) {
            case 1:
                b = [2, 3, 13, 14];
                break;
            case 2:
                b = [1, 3, 14, 13];
                break;
            case 3:
                b = [1, 2];
                break;
            case 4:
                b = [8, 7, 3, 13, 14];
                break;
            case 5:
                b = [7, 8, 1, 14, 13];
                break;
            case 6:
                b = [9, 8, 4, 3];
                break;
            case 7:
                b = [5, 4, 2, 14, 13];
                break;
            case 8:
                b = [4, 5, 3, 13, 14];
                break;
            case 9:
                b = [6, 4, 8, 3];
                break;
            case 10:
                b = [11, 12, 2, 1, 3, 13, 14];
                break;
            case 11:
                b = [10, 12, 1, 2, 3, 14, 13];
                break;
            case 12:
                b = [10, 11];
                break;
            case 13: break;
            case 14: break;
            case 15: break;
        }
        return b;
    };
    DataToolTipLayer.prototype.abl = function (a, b, c, d) {
        return !(this.abi(a, b, c) || this.abj(a, b, c, d));
    };
    DataToolTipLayer.prototype.abm = function (a, b, c, d) {
        return !(this.abk(a, b, c) || this.abh(a, b, c, d));
    };
    DataToolTipLayer.prototype.aff = function (a) {
        var b = new List$1(DataTooltipCollisionInfo.$, 0);
        for (var c = 0; c < a.f.count; c++) {
            var d = new DataTooltipCollisionInfo();
            d.d = c;
            d.e = { $type: Point_$type, x: a.f._inner[c], y: a.g._inner[c] };
            d.c = a.k._inner[c];
            d.a = a.j._inner[c];
            d.b = a.m._inner[c];
            if (isNaN_(d.e.x) || isNaN_(d.e.y)) {
                continue;
            }
            b.add(d);
        }
        b.aa(function (e, f) {
            if (e.e.y < f.e.y) {
                return -1;
            }
            if (e.e.y > f.e.y) {
                return 1;
            }
            if (e.b < f.b) {
                return -1;
            }
            if (e.b > f.b) {
                return 1;
            }
            return 0;
        });
        var e = this.abe(b);
        if (!e) {
            return;
        }
        for (var f = 0; f < b.count - 1; f++) {
            var g = b._inner[f];
            var h = b._inner[f + 1];
            var i = new Rect(0, g.e.x, g.e.y, g.c, g.a);
            var j = new Rect(0, h.e.x, h.e.y, h.c, h.a);
            if (i.intersectsWith(j) || i.top > j.bottom) {
                h.e = { $type: Point_$type, x: h.e.x, y: i.bottom + 1 };
            }
        }
        if (b._inner[b.count - 1].e.y + b._inner[b.count - 1].a > this.cv.bx.height) {
            b._inner[b.count - 1].e = { $type: Point_$type, x: b._inner[b.count - 1].e.x, y: b._inner[b.count - 1].e.y - ((b._inner[b.count - 1].e.y + b._inner[b.count - 1].a) - this.cv.bx.height) };
        }
        for (var k = b.count - 1; k >= 1; k--) {
            var l = b._inner[k];
            var m = b._inner[k - 1];
            var n = new Rect(0, l.e.x, l.e.y, l.c, l.a);
            var o = new Rect(0, m.e.x, m.e.y, m.c, m.a);
            if (n.intersectsWith(o) || n.top < o.bottom) {
                m.e = { $type: Point_$type, x: m.e.x, y: n.top - (o.height + 1) };
            }
        }
        for (var p = 0; p < b.count; p++) {
            var q = b._inner[p].d;
            var r = b._inner[p];
            a.f._inner[q] = r.e.x;
            a.g._inner[q] = r.e.y;
        }
    };
    DataToolTipLayer.prototype.abe = function (a) {
        if (a.count <= 1) {
            return false;
        }
        var b = this.cv.bx;
        for (var c = 0; c < a.count - 1; c++) {
            var d = a._inner[c];
            var e = a._inner[c + 1];
            var f = new Rect(0, d.e.x, d.e.y, d.c, d.a);
            var g = new Rect(0, e.e.x, e.e.y, e.c, e.a);
            if (f.intersectsWith(g)) {
                return true;
            }
        }
        return false;
    };
    DataToolTipLayer.prototype.afj = function (a, b, c, d, e) {
        var _this = this;
        var f = this.yk(a);
        if (!f.c) {
            return;
        }
        var g = a.wc(f.d, c, d);
        var h = this.aar.b9;
        var i = this.zu;
        if (!c) {
            i = this.wh(g);
        }
        var j = a.kn(i);
        var k = h.item(e);
        var l = this.cv.bx;
        var m = this.aar.b8(k);
        if (!Base.equalsStatic(m.target, this.db)) {
            m.target = this.db;
        }
        var n = new List$1(String_$type, 0);
        if (a.eg && this.aao == 2) {
            a.p1(function (o) {
                if (_this.xi(o)) {
                    n.add(o.dataLegendKey);
                }
                return true;
            });
        }
        m.includedSeries = [a.dataLegendKey];
        m.excludedSeries = n.toArray();
        this.afh(m);
        m.targetCursorPositionX = this.zu.x;
        m.targetCursorPositionY = this.zu.y;
        if (!this.aar.cb(k, a)) {
            return;
        }
        var o = this.aar.cq(k);
        b.i.add1(k);
        var p = g.x;
        var q = g.x;
        q = Math.max(q, l.left);
        q = Math.min(q, l.right);
        b.l.add(q);
        var r = g.y;
        r = Math.max(r, l.top);
        r = Math.min(r, l.bottom);
        b.m.add(r);
        var s = DeviceUtils.g(10);
        var t = g.x + s;
        if (t + o.width > l.right) {
            t = (l.right - o.width) - s;
        }
        if (t < l.left) {
            t = l.left;
        }
        b.f.add(t);
        var u = (g.y - o.height) - s;
        if (u + o.height > l.bottom) {
            u = (l.bottom - o.height) - s;
        }
        if (u < l.top) {
            u = l.top;
        }
        b.g.add(u);
        b.k.add(o.width);
        b.j.add(o.height);
        var v = this.u9;
        if (v == null) {
            v = this.vf;
        }
        if (v == null) {
            v = a.u9;
        }
        b.h.add(v);
    };
    DataToolTipLayer.prototype.afi = function (a, b, c) {
        var e_6, _a;
        this.aaf = this.aai();
        this.aaj = this.aam();
        var d = a.f;
        var e = a.e;
        var f = 0;
        var g = this.aar.b9.item(0);
        this.aar.cb(g, null);
        var h = this.aar.b8(g);
        if (!Base.equalsStatic(h.target, this.db)) {
            h.target = this.db;
        }
        var i = new List$1(String_$type, 0);
        var j = this.z5();
        if (j != null) {
            try {
                for (var _b = __values(fromEnum(j.db())), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var k = _c.value;
                    if (k.isStacked) {
                        k.p3(function (l) {
                            i.add(l.dataLegendKey);
                            if (l.eg) {
                                l.p1(function (m) {
                                    i.add(m.dataLegendKey);
                                    return true;
                                });
                            }
                            return true;
                        });
                    }
                    else if (k.eg) {
                        i.add(k.dataLegendKey);
                        k.p1(function (l) {
                            i.add(l.dataLegendKey);
                            return true;
                        });
                    }
                    else {
                        i.add(k.dataLegendKey);
                    }
                }
            }
            catch (e_6_1) {
                e_6 = { error: e_6_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_6)
                        throw e_6.error;
                }
            }
            h.includedSeries = i.toArray();
            h.excludedSeries = null;
        }
        else {
            h.includedSeries = null;
            h.excludedSeries = null;
        }
        this.afh(h);
        h.targetCursorPositionX = this.zu.x;
        h.targetCursorPositionY = this.zu.y;
        var l = this.aar.cq(g);
        var m = 0;
        var n = 0;
        var o = 0;
        var p = 0;
        var q = false;
        var r = false;
        o = d.x;
        p = d.y;
        var s = this.cv.bx;
        var t = this.aar.cr();
        var u = this.ahr();
        var v = this.ahn(u, l, e, d);
        var w = v.x;
        var x = v.y;
        {
        }
        b.l.add(w);
        b.m.add(x);
        b.f.add(w);
        b.g.add(x);
        b.k.add(l.width);
        b.j.add(l.height);
        b.i.add1(g);
        var y = null;
        b.h.add(y);
        this.aar.b9.count = 1;
    };
    DataToolTipLayer.prototype.acz = function (a) {
        return a.toolTip;
    };
    DataToolTipLayer.prototype.yu = function (a, b) {
        _super.prototype.yu.call(this, a, b);
        var c = a;
        var d = c.m.count;
        var e = b.bx.left;
        var f = b.bx.right;
        var g = b.bx.top;
        var h = b.bx.bottom;
        for (var i = 0; i < d; i++) {
            if (isNaN_(c.l._inner[i]) || isNaN_(c.m._inner[i]) || isNaN_(c.f._inner[i]) || isNaN_(c.g._inner[i])) {
                this.aar.ci(c.i._inner[i]);
                continue;
            }
            var j = c.f._inner[i];
            var k = c.g._inner[i];
            var l = c.k._inner[i];
            var m = c.j._inner[i];
            var n = false;
            if (n) {
                this.aar.ci(c.i._inner[i]);
                continue;
            }
            var o = c.i._inner[i];
            var p = Math.min(c.f._inner[i], c.l._inner[i]);
            var q = Math.min(c.g._inner[i], c.m._inner[i]);
            var r = c.l._inner[i] - c.f._inner[i];
            var s = c.m._inner[i] - c.g._inner[i];
            if (o != null) {
                this.aar.ck(o, p, q, r, s);
            }
        }
    };
    DataToolTipLayer.prototype.ga = function () {
        return true;
    };
    Object.defineProperty(DataToolTipLayer.prototype, "aag", {
        get: function () {
            return this.aaf;
        },
        set: function (a) {
            var b = this.aaf;
            this.aaf = a;
            if (b != this.aaf) {
                this.raisePropertyChanged("ActualGroupedPositionModeX", enumGetBox(DataTooltipGroupedPositionX_$type, b), enumGetBox(DataTooltipGroupedPositionX_$type, this.aaf));
            }
        },
        enumerable: false,
        configurable: true
    });
    DataToolTipLayer.prototype.aai = function () {
        var e_7, _a;
        if (this.db == null) {
            return 2;
        }
        var a = false;
        var b = false;
        var c = false;
        var d = false;
        var e = this.z5();
        if (e == null) {
            if (this.aah != 0) {
                return this.aah;
            }
            else {
                return 2;
            }
        }
        try {
            for (var _b = __values(fromEnum(e.df)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var f = _c.value;
                if (f.ez) {
                    a = true;
                }
                if (f.isVertical) {
                    b = true;
                }
                if (f.fp) {
                    c = true;
                }
                if (f.fw) {
                    d = true;
                }
            }
        }
        catch (e_7_1) {
            e_7 = { error: e_7_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_7)
                    throw e_7.error;
            }
        }
        var g = ((function () {
            var $ret = new HashSet$1(DataTooltipGroupedPositionX_$type, 0);
            $ret.add_1(1);
            $ret.add_1(2);
            $ret.add_1(3);
            $ret.add_1(4);
            $ret.add_1(5);
            $ret.add_1(6);
            $ret.add_1(7);
            $ret.add_1(8);
            $ret.add_1(9);
            return $ret;
        })());
        var h = g.contains(this.aah);
        if (this.aah != 0) {
            return this.aah;
        }
        if (h && b) {
            return 11;
        }
        if (c || d) {
            return 11;
        }
        if (b) {
            return 11;
        }
        if (a) {
            return 8;
        }
        return 8;
    };
    Object.defineProperty(DataToolTipLayer.prototype, "aak", {
        get: function () {
            return this.aaj;
        },
        set: function (a) {
            var b = this.aaj;
            this.aaj = a;
            if (b != this.aaj) {
                this.raisePropertyChanged("ActualPositionModeY", enumGetBox(DataTooltipGroupedPositionY_$type, b), enumGetBox(DataTooltipGroupedPositionY_$type, this.aaj));
            }
        },
        enumerable: false,
        configurable: true
    });
    DataToolTipLayer.prototype.aam = function () {
        var e_8, _a;
        if (this.db == null) {
            return 11;
        }
        var a = false;
        var b = false;
        var c = false;
        var d = this.z5();
        if (d == null) {
            if (this.aal != 0) {
                return this.aal;
            }
            else {
                return 11;
            }
        }
        try {
            for (var _b = __values(fromEnum(d.db())), _c = _b.next(); !_c.done; _c = _b.next()) {
                var e = _c.value;
                if (e.isVertical) {
                    a = true;
                }
                if (e.fp) {
                    b = true;
                }
                if (e.fw) {
                    c = true;
                }
            }
        }
        catch (e_8_1) {
            e_8 = { error: e_8_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_8)
                    throw e_8.error;
            }
        }
        var f = ((function () {
            var $ret = new HashSet$1(DataTooltipGroupedPositionY_$type, 0);
            $ret.add_1(1);
            $ret.add_1(2);
            $ret.add_1(3);
            $ret.add_1(4);
            $ret.add_1(5);
            $ret.add_1(6);
            $ret.add_1(7);
            $ret.add_1(8);
            $ret.add_1(9);
            return $ret;
        })());
        var g = f.contains(this.aal);
        if (this.aal != 0) {
            return this.aal;
        }
        if (g && !a) {
            return 11;
        }
        if (b || c) {
            return 11;
        }
        if (a) {
            return 8;
        }
        if (!a) {
            return 11;
        }
        return 11;
    };
    Object.defineProperty(DataToolTipLayer.prototype, "abd", {
        get: function () {
            return this.aa3;
        },
        set: function (a) {
            var b = this.aa3;
            this.aa3 = a;
            if (b != this.aa3) {
                this.raisePropertyChanged("ValueTextStyle", b, this.aa3);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "aa5", {
        get: function () {
            return this.aav;
        },
        set: function (a) {
            var b = this.aav;
            this.aav = a;
            if (b != this.aav) {
                this.raisePropertyChanged("HeaderTextStyle", b, this.aav);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "aa4", {
        get: function () {
            return this.aau;
        },
        set: function (a) {
            var b = this.aau;
            this.aau = a;
            if (b != this.aau) {
                this.raisePropertyChanged("GroupTextStyle", b, this.aau);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "aa8", {
        get: function () {
            return this.aay;
        },
        set: function (a) {
            var b = this.aay;
            this.aay = a;
            if (b != this.aay) {
                this.raisePropertyChanged("SummaryTitleTextStyle", b, this.aay);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "aba", {
        get: function () {
            return this.aa0;
        },
        set: function (a) {
            var b = this.aa0;
            this.aa0 = a;
            if (b != this.aa0) {
                this.raisePropertyChanged("SummaryValueTextStyle", b, this.aa0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "aa7", {
        get: function () {
            return this.aax;
        },
        set: function (a) {
            var b = this.aax;
            this.aax = a;
            if (b != this.aax) {
                this.raisePropertyChanged("SummaryLabelTextStyle", b, this.aax);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "aa9", {
        get: function () {
            return this.aaz;
        },
        set: function (a) {
            var b = this.aaz;
            this.aaz = a;
            if (b != this.aaz) {
                this.raisePropertyChanged("SummaryUnitsTextStyle", b, this.aaz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "abc", {
        get: function () {
            return this.aa2;
        },
        set: function (a) {
            var b = this.aa2;
            this.aa2 = a;
            if (b != this.aa2) {
                this.raisePropertyChanged("UnitsTextStyle", b, this.aa2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "abb", {
        get: function () {
            return this.aa1;
        },
        set: function (a) {
            var b = this.aa1;
            this.aa1 = a;
            if (b != this.aa1) {
                this.raisePropertyChanged("TitleTextStyle", b, this.aa1);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataToolTipLayer.prototype, "aa6", {
        get: function () {
            return this.aaw;
        },
        set: function (a) {
            var b = this.aaw;
            this.aaw = a;
            if (b != this.aaw) {
                this.raisePropertyChanged("LabelTextStyle", b, this.aaw);
            }
        },
        enumerable: false,
        configurable: true
    });
    DataToolTipLayer.$t = markType(DataToolTipLayer, 'DataToolTipLayer', AnnotationLayer.$);
    DataToolTipLayer.$$p = markDep(DependencyProperty, PropertyMetadata, DataToolTipLayer, 'raisePropertyChanged', ['BadgeMargin:aht:afl', [Thickness.$, new Thickness(1, 0, 0, 0, 0)], 'BadgeShape:aat:afm', [LegendItemBadgeShape_$type, enumGetBox(LegendItemBadgeShape_$type, 0)], 'DefaultPositionOffsetX:abv:afn', [1, DeviceUtils.g(10)], 'DefaultPositionOffsetY:abw:afo', [1, DeviceUtils.g(10)], 'ExcludedColumns:zx:afp', [Array_$type, null], 'ExcludedSeries:zy:afq', [Array_$type, null], 'GroupRowMargin:ahu:afu', [Thickness.$, new Thickness(0, 5)], 'GroupRowVisible:abf:afv', [0, false], 'GroupTextColor:ahd:afw', [Brush.$, null], 'GroupTextMargin:ahv:afx', [Thickness.$, new Thickness(1, 2, 0, 0, 0)], 'GroupedPositionModeX:aah:afr', [DataTooltipGroupedPositionX_$type, enumGetBox(DataTooltipGroupedPositionX_$type, 0)], 'GroupedPositionModeY:aal:afs', [DataTooltipGroupedPositionY_$type, enumGetBox(DataTooltipGroupedPositionY_$type, 0)], 'GroupingMode:aap:aft', [DataToolTipLayerGroupingMode_$type, enumGetBox(DataToolTipLayerGroupingMode_$type, 0)], 'HeaderFormatCulture:adh:afy', [2, stringEmpty()], 'HeaderFormatDate:z8:afz', [DataLegendHeaderDateMode_$type, enumGetBox(DataLegendHeaderDateMode_$type, 0)], 'HeaderFormatSpecifiers:z1:af0', [Array_$type, null], 'HeaderFormatString:adl:af1', [2, null], 'HeaderFormatTime:z9:af2', [DataLegendHeaderTimeMode_$type, enumGetBox(DataLegendHeaderTimeMode_$type, 0)], 'HeaderRowMargin:ahw:af3', [Thickness.$, new Thickness(0, 5)], 'HeaderRowVisible:abg:af4', [0, true], 'HeaderTextColor:ahe:af5', [Brush.$, null], 'HeaderTextMargin:ahx:af6', [Thickness.$, new Thickness(1, 2, 0, 0, 0)], 'HeaderText:adq:af7', [2, null], 'IncludedColumns:z2:af8', [Array_$type, null], 'IncludedSeries:z3:af9', [Array_$type, null], 'LabelDisplayMode:aaa:aga', [DataLegendLabelMode_$type, enumGetBox(DataLegendLabelMode_$type, 0)], 'LabelTextColor:ahf:agb', [Brush.$, null], 'LabelTextMargin:ahy:agc', [Thickness.$, new Thickness(1, 8, 0, 0, 0)], 'LayoutMode:aab:agd', [DataLegendLayoutMode_$type, enumGetBox(DataLegendLayoutMode_$type, 0)], 'PositionOffsetX:abx:age', [1, NaN], 'PositionOffsetY:aby:agf', [1, NaN], 'ShouldUpdateWhenSeriesDataChanges:abn:agg', [0, true], 'SummaryLabelTextColor:ahg:agh', [Brush.$, null], 'SummaryLabelText:ad5:agi', [2, null], 'SummaryRowMargin:ahz:agj', [Thickness.$, new Thickness(0, 5)], 'SummaryTitleTextColor:ahh:agk', [Brush.$, null], 'SummaryTitleTextMargin:ah0:agl', [Thickness.$, new Thickness(1, 2, 0, 0, 0)], 'SummaryTitleText:aea:agm', [2, null], 'SummaryType:aac:agn', [DataLegendSummaryType_$type, enumGetBox(DataLegendSummaryType_$type, 0)], 'SummaryUnitsTextColor:ahi:ago', [Brush.$, null], 'SummaryUnitsText:aeg:agp', [2, null], 'SummaryValueTextColor:ahj:agq', [Brush.$, null], 'TargetAxis:z6:agr', [Axis.$, null], 'TitleTextColor:ahk:ags', [Brush.$, null], 'TitleTextMargin:ah1:agt', [Thickness.$, new Thickness(1, 0, 0, 0, 0)], 'UnitsDisplayMode:aad:agu', [DataLegendUnitsMode_$type, enumGetBox(DataLegendUnitsMode_$type, 0)], 'UnitsTextColor:ahl:agv', [Brush.$, null], 'UnitsTextMargin:ah2:agw', [Thickness.$, new Thickness(1, 0, 0, 0, 0)], 'UnitsText:aer:agx', [2, null], 'UseInterpolation:abo:agy', [0, false], 'ValueFormatAbbreviation:z7:agz', [DataAbbreviationMode_$type, enumGetBox(DataAbbreviationMode_$type, 0)], 'ValueFormatCulture:aey:ag0', [2, stringEmpty()], 'ValueFormatMaxFractions:acx:ag1', [1, -1], 'ValueFormatMinFractions:acy:ag2', [1, -1], 'ValueFormatMode:aae:ag3', [DataLegendValueMode_$type, enumGetBox(DataLegendValueMode_$type, 0)], 'ValueFormatSpecifiers:z4:ag4', [Array_$type, null], 'ValueFormatString:ae4:ag5', [2, null], 'ValueFormatUseGrouping:abp:ag6', [0, true], 'ValueRowMargin:ah3:ag7', [Thickness.$, new Thickness(0, 5)], 'ValueRowVisible:abq:ag8', [0, true], 'ValueTextColor:ahm:ag9', [Brush.$, null], 'ValueTextMargin:ah4:aha', [Thickness.$, new Thickness(1, 2, 0, 2, 0)], 'ValueTextUseSeriesColors:abr:ahb', [0, false], 'ValueTextWhenMissingData:afd:ahc', [2, "N/A"]]);
    return DataToolTipLayer;
}(AnnotationLayer));
export { DataToolTipLayer };
/**
 * @hidden
 */
var DataToolTipLayerView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataToolTipLayerView, _super);
    function DataToolTipLayerView(a) {
        var _this = _super.call(this, a) || this;
        _this.b9 = null;
        _this.cc = null;
        _this.b7 = null;
        _this.ca = false;
        _this._htmlTest = null;
        _this.b7 = a;
        _this.htmlTest = /^[^<]*(<[\w\W]+>)[^>]*$/;
        _this.cc = new List$1(PointerTooltip.$, 0);
        return _this;
    }
    DataToolTipLayerView.prototype.ak = function () {
        _super.prototype.ak.call(this);
        for (var a = 0; a < this.b9.a.count; a++) {
            this.b9.a._inner[a].br();
        }
        for (var b = 0; b < this.b9.b.count; b++) {
            this.b9.b._inner[b].br();
        }
        this.b9.clear();
    };
    DataToolTipLayerView.prototype.cd = function () {
        var a = new PointerTooltip();
        a.bw(this.e.jz);
        a._visibility = 1;
        a.ap = this.e.db.c9.k;
        var b = null;
        if (TypeRegistrar.isRegistered("XamDataLegend")) {
            b = TypeRegistrar.create("XamDataLegend");
        }
        b.pixelScalingRatio = 1;
        if (this.e.db != null) {
            var c = this.e.db.c9.j.ae.createElement("div");
            var ren_ = this.e.db.c9.j.ae.getSubRenderer(c);
            var len_ = b;
            len_.provideContainer(ren_);
        }
        a.ai = typeCast(IOnDemandRender_$type, b);
        a.ai.shouldOnDemandRender = true;
        this.cc.add(a);
        return a;
    };
    DataToolTipLayerView.prototype.cf = function (a) {
        var b = a;
    };
    DataToolTipLayerView.prototype.ch = function (a) {
        var b = a;
        b._visibility = 1;
    };
    DataToolTipLayerView.prototype.cg = function (a) {
        var b = a;
        b.bz();
        b.content = null;
        this.cc.remove1(a);
    };
    DataToolTipLayerView.prototype.au = function () {
        var _this = this;
        _super.prototype.au.call(this);
        this.b9 = ((function () {
            var $ret = new Pool$1(Base.$);
            $ret.create = runOn(_this, _this.cd);
            $ret.activate = runOn(_this, _this.cf);
            $ret.disactivate = runOn(_this, _this.ch);
            $ret.destroy = runOn(_this, _this.cg);
            return $ret;
        })());
    };
    Object.defineProperty(DataToolTipLayerView.prototype, "htmlTest", {
        get: function () {
            return this._htmlTest;
        },
        set: function (a) {
            this._htmlTest = a;
        },
        enumerable: false,
        configurable: true
    });
    DataToolTipLayerView.prototype.cb = function (a, b) {
        var c = a;
        var d = c.ai;
        return true;
    };
    DataToolTipLayerView.prototype.cq = function (a) {
        var b = a;
        var c = new Size(1, Number.POSITIVE_INFINITY, Number.POSITIVE_INFINITY);
        b.cy = 1;
        var d = b.cx(new Size(1, Number.POSITIVE_INFINITY, Number.POSITIVE_INFINITY));
        b.cy = 0;
        return d;
    };
    DataToolTipLayerView.prototype.cp = function () {
        if (this.e.db != null) {
            var a = this.e.db.vx();
            var b = this.e.db.c9.a8;
            var c = this.e.db.c9.a7;
            return new Rect(0, 0, 0, b, c);
        }
        return Rect.empty;
    };
    DataToolTipLayerView.prototype.co = function () {
        if (this.e.db != null) {
            var a = this.e.db.vx();
            var b = this.e.db.c9.a8;
            var c = this.e.db.c9.a7;
            return new Rect(0, 0, 0, b, c);
        }
        return Rect.empty;
    };
    DataToolTipLayerView.prototype.cn = function () {
        var a = 0;
        var b = 0;
        if (this.e.db != null) {
            var c = this.e.db.vx();
            a += c.x;
            b += c.y;
            var d = this.e.db.c9.a8;
            var e = this.e.db.c9.a7;
            return new Rect(0, a, b, d, e);
        }
        return Rect.empty;
    };
    DataToolTipLayerView.prototype.cm = function (a) {
        var b = 0;
        var c = 0;
        if (this.e.db != null) {
            var d = this.e.db.vx();
            b += d.x;
            c += d.y;
        }
        return { $type: Point_$type, x: b + a.x, y: c + a.y };
    };
    DataToolTipLayerView.prototype.ck = function (a, b, c, d, e) {
        var f = null;
        if (this.e.db != null) {
            f = this.e.db.c9.j.i();
        }
        var g = a;
        if (g == null) {
            return;
        }
        if (d == 0 && e == 0) {
            g.cy = 1;
        }
        else {
            g.cy = 0;
        }
        g.aq = f;
        g.cs = { $type: Point_$type, x: d, y: e };
        g._visibility = 0;
        g.n = b;
        g.o = c;
    };
    DataToolTipLayerView.prototype.cr = function () {
        var a = (window.innerWidth);
        var b = (window.innerHeight);
        return new Size(1, a, b);
    };
    DataToolTipLayerView.prototype.ci = function (a) {
        if (a != null) {
            var b = a;
            b._visibility = 1;
        }
    };
    DataToolTipLayerView.prototype.cl = function () {
        if (this.e.db == null) {
            this.cj(true);
        }
    };
    DataToolTipLayerView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (b) {
            return;
        }
        var c = 0;
        var d = 0;
        var e = this.e.db;
        if (e != null) {
            var f = e.c9.j.b8();
            c += f.x;
            d += f.y;
        }
        for (var g = 0; g < this.cc.count; g++) {
            this.cc._inner[g].by(this.cc._inner[g].n + c, this.cc._inner[g].o + d);
        }
    };
    DataToolTipLayerView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        for (var b = 0; b < this.cc.count; b++) {
            var c = this.cc._inner[b];
            var d = 0;
            var e = 0;
            if (this.e.db != null) {
                var f = this.e.db.vx();
                d += f.x;
                e += f.y;
            }
            var g = c.a5();
            var h = c.n - d;
            var i = c.o - e;
            var j = c.ai;
            var k = null;
            if (j != null) {
                k = j.exportVisualData();
            }
            this.e.bc.addPointerTooltipWithDataLegend(a, g, h, i, k);
        }
    };
    DataToolTipLayerView.prototype.b8 = function (a) {
        var b = a;
        return b.ai;
    };
    DataToolTipLayerView.prototype.cj = function (a) {
        this.as();
        for (var b = 0; b < this.cc.count; b++) {
            this.ci(this.cc._inner[b]);
            if (a) {
                this.cc._inner[b].bs();
            }
        }
    };
    DataToolTipLayerView.$t = markType(DataToolTipLayerView, 'DataToolTipLayerView', AnnotationLayerView.$);
    return DataToolTipLayerView;
}(AnnotationLayerView));
export { DataToolTipLayerView };
