// igniteui-react-charts/esm5/lib/DataUpdatedMessage.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ConfigurationMessage } from "./ConfigurationMessage";
import { ItemsSourceAction_$type } from "./ItemsSourceAction";
import { EnumUtil, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var DataUpdatedMessage = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataUpdatedMessage, _super);
    function DataUpdatedMessage() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.b = 0;
        _this.e = 0;
        _this.d = 0;
        _this.h = null;
        return _this;
    }
    Object.defineProperty(DataUpdatedMessage.prototype, "c", {
        get: function () {
            return this.b;
        },
        set: function (a) {
            this.b = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataUpdatedMessage.prototype, "g", {
        get: function () {
            return this.e;
        },
        set: function (a) {
            this.e = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataUpdatedMessage.prototype, "f", {
        get: function () {
            return this.d;
        },
        set: function (a) {
            this.d = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataUpdatedMessage.prototype, "i", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
        },
        enumerable: false,
        configurable: true
    });
    DataUpdatedMessage.prototype.toString = function () {
        return "DataUpdatedMessage[" + EnumUtil.getName(ItemsSourceAction_$type, this.c) + "," + this.g.toString() + ", " + this.f.toString() + ", " + this.i.toString() + "]";
    };
    DataUpdatedMessage.$t = markType(DataUpdatedMessage, 'DataUpdatedMessage', ConfigurationMessage.$);
    return DataUpdatedMessage;
}(ConfigurationMessage));
export { DataUpdatedMessage };
