// igniteui-react-charts/esm5/lib/DefaultProvidesViewport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, markType } from "igniteui-react-core";
import { IProvidesViewport_$type } from "./IProvidesViewport";
import { Rect } from "igniteui-react-core";
/**
 * @hidden
 */
var DefaultProvidesViewport = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DefaultProvidesViewport, _super);
    function DefaultProvidesViewport() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DefaultProvidesViewport.prototype.getViewInfo = function (a, b) {
        a = Rect.empty;
        b = Rect.empty;
        return {
            p0: a,
            p1: b
        };
    };
    DefaultProvidesViewport.prototype.getEffectiveViewport = function () {
        return Rect.empty;
    };
    DefaultProvidesViewport.prototype.getMainContentViewport = function () {
        return Rect.empty;
    };
    DefaultProvidesViewport.$t = markType(DefaultProvidesViewport, 'DefaultProvidesViewport', Base.$, [IProvidesViewport_$type]);
    return DefaultProvidesViewport;
}(Base));
export { DefaultProvidesViewport };
