// igniteui-react-charts/esm5/lib/DetrendedPriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { DetrendedPriceOscillatorIndicatorStrategy } from "./DetrendedPriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var DetrendedPriceOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DetrendedPriceOscillatorIndicator, _super);
    function DetrendedPriceOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DetrendedPriceOscillatorIndicator.prototype.bj = function () {
        return new DetrendedPriceOscillatorIndicator();
    };
    DetrendedPriceOscillatorIndicator.prototype.get_abv = function () {
        return new DetrendedPriceOscillatorIndicatorStrategy();
    };
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    DetrendedPriceOscillatorIndicator.prototype.get_ab8 = function () {
        return DetrendedPriceOscillatorIndicator.$;
    };
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(DetrendedPriceOscillatorIndicator.ach));
        },
        set: function (a) {
            this.h(DetrendedPriceOscillatorIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    DetrendedPriceOscillatorIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    DetrendedPriceOscillatorIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    DetrendedPriceOscillatorIndicator.$t = markType(DetrendedPriceOscillatorIndicator, 'DetrendedPriceOscillatorIndicator', StrategyBasedIndicator.$);
    DetrendedPriceOscillatorIndicator.ach = StrategyBasedIndicator.acd(20, DetrendedPriceOscillatorIndicator.$);
    return DetrendedPriceOscillatorIndicator;
}(StrategyBasedIndicator));
export { DetrendedPriceOscillatorIndicator };
