// igniteui-react-charts/esm5/lib/DomainType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * An enum representing the available types of domain charts.
 */
export var DomainType = /*@__PURE__*/ (function (DomainType) {
    /**
     * Specifies category domain for Category Chart
     */
    DomainType[DomainType["Category"] = 0] = "Category";
    /**
     * Specifies financial domain for Financial Chart
     */
    DomainType[DomainType["Financial"] = 1] = "Financial";
    /**
     * Specifies scatter domain for Scatter Charts
     */
    DomainType[DomainType["Scatter"] = 2] = "Scatter";
    /**
     * Specifies shape domain for Shape Charts
     */
    DomainType[DomainType["Shape"] = 3] = "Shape";
    /**
     * Specifies pie domain for Pie charts
     */
    DomainType[DomainType["Pie"] = 4] = "Pie";
    return DomainType;
})({});
/**
 * @hidden
 */
export var DomainType_$type = /*@__PURE__*/ markEnum('DomainType', 'Category,0|Financial,1|Scatter,2|Shape,3|Pie,4');
