// igniteui-react-charts/esm5/lib/EnableErrorBars.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available types for enabling error bars
 */
export var EnableErrorBars = /*@__PURE__*/ (function (EnableErrorBars) {
    /**
     * No error bars are enabled
     */
    EnableErrorBars[EnableErrorBars["None"] = 0] = "None";
    /**
     * Both error bars are enabled
     */
    EnableErrorBars[EnableErrorBars["Both"] = 1] = "Both";
    /**
     * Positive error bars are enabled
     */
    EnableErrorBars[EnableErrorBars["Positive"] = 2] = "Positive";
    /**
     * Negative error bars are enabled
     */
    EnableErrorBars[EnableErrorBars["Negative"] = 3] = "Negative";
    return EnableErrorBars;
})({});
/**
 * @hidden
 */
export var EnableErrorBars_$type = /*@__PURE__*/ markEnum('EnableErrorBars', 'None,0|Both,1|Positive,2|Negative,3');
