// igniteui-react-charts/esm5/lib/FinancialEventArgs_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __generator, __values } from "tslib";
import { EventArgs, Base, markType, typeCast, runOn, delegateCombine, enumGetBox, Point_$type, String_$type, Number_$type, fromEnum, toEnum, Date_$type, markDep, Delegate_$type } from "igniteui-react-core";
import { IBucketizer_$type } from "./IBucketizer";
import { ArgumentNullException } from "igniteui-react-core";
import { IFastItemColumn$1_$type } from "igniteui-react-core";
import { Series } from "./Series";
import { Rect } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { ISortingAxis_$type } from "./ISortingAxis";
import { SeriesView } from "./SeriesView";
import { truncate, intDivide, isInfinity, isNaN_ } from "igniteui-react-core";
import { IHasCategoryAxis_$type } from "./IHasCategoryAxis";
import { IHasCategoryModePreference_$type } from "./IHasCategoryModePreference";
import { Brush } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { NumericYAxis } from "./NumericYAxis";
import { HighlightingInfo } from "igniteui-react-core";
import { CategoryTransitionInMode_$type } from "./CategoryTransitionInMode";
import { CategoryTransitionSourceFramePreparer } from "./CategoryTransitionSourceFramePreparer";
import { CategoryFrame } from "./CategoryFrame";
import { CategoryMode_$type } from "./CategoryMode";
import { ITimeRangeSelectorAxis_$type } from "./ITimeRangeSelectorAxis";
import { FastReflectionHelper } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { SeriesLayerPropertyOverlay } from "./SeriesLayerPropertyOverlay";
import { Dictionary$2 } from "igniteui-react-core";
import { CategorySeriesRenderManager } from "./CategorySeriesRenderManager";
import { IFastItemsSource_$type } from "igniteui-react-core";
import { FinancialValueList } from "./FinancialValueList";
import { CategoryXAxis } from "./CategoryXAxis";
import { SafeSortedReadOnlyDoubleCollection } from "./SafeSortedReadOnlyDoubleCollection";
import { SafeReadOnlyDoubleCollection } from "./SafeReadOnlyDoubleCollection";
import { SafeEnumerable } from "./SafeEnumerable";
import { CalculatedColumn } from "./CalculatedColumn";
import { TrendCalculators } from "igniteui-react-core";
import { ColumnSupportingCalculation } from "./ColumnSupportingCalculation";
import { IDateTimeAxis_$type } from "./IDateTimeAxis";
import { PropertyMetadata } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
import { dateToStringFormat } from "igniteui-react-core";
import { SupportingCalculation$1 } from "./SupportingCalculation$1";
import { SRProvider } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
var FinancialEventArgs = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialEventArgs, _super);
    function FinancialEventArgs(a, b, c, d, e) {
        var _this = _super.call(this) || this;
        _this._position = 0;
        _this._count = 0;
        _this._dataSource = null;
        _this._supportingCalculations = null;
        _this._basedOn = null;
        _this.position = a;
        _this.count = b;
        _this.dataSource = c;
        _this.supportingCalculations = d;
        if (c != null) {
            c.series = e;
        }
        return _this;
    }
    Object.defineProperty(FinancialEventArgs.prototype, "position", {
        get: function () {
            return this._position;
        },
        set: function (a) {
            this._position = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialEventArgs.prototype, "count", {
        get: function () {
            return this._count;
        },
        set: function (a) {
            this._count = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialEventArgs.prototype, "dataSource", {
        get: function () {
            return this._dataSource;
        },
        set: function (a) {
            this._dataSource = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialEventArgs.prototype, "supportingCalculations", {
        get: function () {
            return this._supportingCalculations;
        },
        set: function (a) {
            this._supportingCalculations = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialEventArgs.prototype, "basedOn", {
        get: function () {
            return this._basedOn;
        },
        set: function (a) {
            this._basedOn = a;
        },
        enumerable: false,
        configurable: true
    });
    FinancialEventArgs.$t = markType(FinancialEventArgs, 'FinancialEventArgs', EventArgs.$);
    return FinancialEventArgs;
}(EventArgs));
export { FinancialEventArgs };
/**
 * @hidden
 */
var FinancialBucketCalculator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialBucketCalculator, _super);
    function FinancialBucketCalculator(a) {
        var _this = _super.call(this) || this;
        _this.a = null;
        _this.c = 0;
        _this.d = 0;
        _this.b = 0;
        if (a == null) {
            throw new ArgumentNullException(0, "view");
        }
        _this.a = a;
        _this.c = -1;
        _this.d = _this.d;
        _this.b = 0;
        return _this;
    }
    FinancialBucketCalculator.prototype.getBucket = function (a) {
        return null;
    };
    FinancialBucketCalculator.prototype.getBucketWithoutUnknowns = function (a) {
        return this.getBucket(a);
    };
    FinancialBucketCalculator.prototype.getErrorBucket = function (a, b) {
        return NaN;
    };
    FinancialBucketCalculator.prototype.getBucketInfo = function (a, b, c, d) {
        a = this.c;
        b = this.d;
        c = this.b;
        d = this.a.b2.id;
        return {
            p0: a,
            p1: b,
            p2: c,
            p3: d
        };
    };
    FinancialBucketCalculator.prototype.e = function (a) {
        var b = this.a.by;
        var c = this.a.bx;
        var d = this.a.bx;
        if (this.a.b2 != null) {
            d = this.a.b2.getEffectiveViewport1(this.a);
        }
        if (b.isEmpty || c.isEmpty || this.a.b2.w6 == null) {
            this.b = 0;
            return;
        }
        var e = (this.a.b2.w6 != null) ? this.a.b2.w6.ch : false;
        var f = new ScalerParams(0, b, c, e, d);
        var g = this.a.b2.wj(this.a);
        var h = typeCast(ISortingAxis_$type, this.a.b2.w6);
        if (h == null || h.i$b == null) {
            var i = Math.floor(this.a.b2.w6.eo(g.left, f));
            var j = Math.ceil(this.a.b2.w6.eo(g.right, f));
            if (this.a.b2.w6.ch) {
                j = Math.ceil(this.a.b2.w6.eo(g.left, f));
                i = Math.floor(this.a.b2.w6.eo(g.right, f));
            }
            var k = Math.floor((j - i + 1) * a / g.width);
            this.b = truncate(Math.max(1, k));
            this.c = truncate(Math.floor(i / this.b));
            this.d = truncate(Math.ceil(j / this.b));
        }
        else {
            this.c = h.i$e(b, c, d, g);
            this.d = h.i$g(b, c, d, g);
            this.b = 1;
        }
    };
    FinancialBucketCalculator.prototype.cacheValues = function () {
    };
    FinancialBucketCalculator.prototype.unCacheValues = function () {
    };
    FinancialBucketCalculator.$t = markType(FinancialBucketCalculator, 'FinancialBucketCalculator', Base.$, [IBucketizer_$type]);
    return FinancialBucketCalculator;
}(Base));
export { FinancialBucketCalculator };
/**
 * @hidden
 */
var FinancialSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialSeries, _super);
    function FinancialSeries() {
        var _this = _super.call(this) || this;
        _this.xx = null;
        _this.xv = null;
        _this.xw = null;
        _this.xu = null;
        _this.xy = null;
        _this.x7 = false;
        _this.assigningCategoryStyle = null;
        _this.xf = null;
        _this.yh = new Dictionary$2(IFastItemColumn$1_$type.specialize(Number_$type), String_$type, 0);
        _this.yi = new Dictionary$2(String_$type, String_$type, 0);
        _this.x6 = false;
        _this.xt = null;
        _this.xa = null;
        _this.xc = null;
        _this.w8 = null;
        _this.xi = null;
        _this.typical = null;
        _this.typicalBasedOn = null;
        _this.x5 = false;
        _this.yu = null;
        _this.ys = null;
        _this.yt = null;
        _this.yr = null;
        _this.yv = null;
        _this.xb = null;
        _this.w7 = null;
        _this.xd = 0;
        _this.xo = null;
        _this.xq = null;
        _this.xf = new CategorySeriesRenderManager();
        _this.x7 = false;
        _this.xi = new CategoryTransitionSourceFramePreparer();
        _this.xb = new CategoryFrame(3);
        _this.ab = FinancialSeries.$;
        _this.xe = 0;
        return _this;
    }
    FinancialSeries.prototype.ht = function () {
        var a = _super.prototype.ht.call(this);
        {
            var b = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "OpenMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedOpenMemberPath";
                $ret.q = true;
                return $ret;
            })());
            b.valueResolving = delegateCombine(b.valueResolving, runOn(this, this.z9));
            a.add(b);
        }
        {
            var c = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "HighMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedHighMemberPath";
                $ret.q = true;
                return $ret;
            })());
            c.valueResolving = delegateCombine(c.valueResolving, runOn(this, this.z7));
            a.add(c);
        }
        {
            var d = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "LowMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedLowMemberPath";
                $ret.q = true;
                return $ret;
            })());
            d.valueResolving = delegateCombine(d.valueResolving, runOn(this, this.z8));
            a.add(d);
        }
        {
            var e = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "CloseMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedCloseMemberPath";
                $ret.q = true;
                return $ret;
            })());
            e.valueResolving = delegateCombine(e.valueResolving, runOn(this, this.z6));
            a.add(e);
        }
        {
            var f = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "VolumeMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedVolumeMemberPath";
                $ret.q = true;
                return $ret;
            })());
            f.valueResolving = delegateCombine(f.valueResolving, runOn(this, this.aaa));
            a.add(f);
        }
        {
            var g = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Brush";
                $ret.p = true;
                $ret.y = "ActualBrush";
                $ret.q = true;
                return $ret;
            })());
            g.valueResolving = delegateCombine(g.valueResolving, runOn(this, this.r0));
            a.add(g);
        }
        {
            var h = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Outline";
                $ret.p = true;
                $ret.y = "ActualOutline";
                $ret.q = true;
                return $ret;
            })());
            h.valueResolving = delegateCombine(h.valueResolving, runOn(this, this.r0));
            a.add(h);
        }
        {
            var i = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "NegativeBrush";
                $ret.p = true;
                $ret.q = true;
                return $ret;
            })());
            i.valueResolving = delegateCombine(i.valueResolving, runOn(this, this.r0));
            a.add(i);
        }
        {
            var j = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "NegativeOutline";
                $ret.p = true;
                $ret.q = true;
                return $ret;
            })());
            j.valueResolving = delegateCombine(j.valueResolving, runOn(this, this.r0));
            a.add(j);
        }
        var k = ((function () {
            var $ret = new SeriesLayerPropertyOverlay();
            $ret.p = true;
            $ret.ad = "IsTransitionInEnabled";
            $ret.x = true;
            $ret.q = true;
            return $ret;
        })());
        a.add(k);
        var l = ((function () {
            var $ret = new SeriesLayerPropertyOverlay();
            $ret.p = true;
            $ret.ad = "TransitionInMode";
            $ret.x = enumGetBox(CategoryTransitionInMode_$type, 20);
            $ret.q = true;
            return $ret;
        })());
        a.add(l);
        return a;
    };
    FinancialSeries.prototype.z9 = function (a, b) {
        if (this.zj != null) {
            b.value = this.zj;
            return;
        }
        b.value = this.zw;
    };
    FinancialSeries.prototype.z7 = function (a, b) {
        if (this.zf != null) {
            b.value = this.zf;
            return;
        }
        b.value = this.zn;
    };
    FinancialSeries.prototype.z8 = function (a, b) {
        if (this.zh != null) {
            b.value = this.zh;
            return;
        }
        b.value = this.zs;
    };
    FinancialSeries.prototype.z6 = function (a, b) {
        if (this.zd != null) {
            b.value = this.zd;
            return;
        }
        b.value = this.za;
    };
    FinancialSeries.prototype.aaa = function (a, b) {
        if (this.zl != null) {
            b.value = this.zl;
            return;
        }
        b.value = this.z0;
    };
    FinancialSeries.prototype.kv = function () {
        if (this.d3()) {
            return this.bn.xc;
        }
        else {
            return this.bn.w8;
        }
    };
    FinancialSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.xn = a;
    };
    FinancialSeries.prototype.get_e6 = function () {
        return true;
    };
    FinancialSeries.prototype.ai = function () {
        return [this.w6, this.xp];
    };
    FinancialSeries.prototype.kq = function (a, b) {
        var c = this.nd(b);
        if (c == FinancialSeries.$$p[11]) {
            return this.kr(a, b, this.zw);
        }
        if (c == FinancialSeries.$$p[1]) {
            return this.kr(a, b, this.zn);
        }
        if (c == FinancialSeries.$$p[9]) {
            return this.kr(a, b, this.zs);
        }
        if (c == FinancialSeries.$$p[0]) {
            return this.kr(a, b, this.za);
        }
        if (c == FinancialSeries.$$p[13]) {
            return this.kr(a, b, this.z0);
        }
        if (c == FinancialSeries.$$p[5]) {
            return this.kr(a, b, this.zj);
        }
        if (c == FinancialSeries.$$p[3]) {
            return this.kr(a, b, this.zf);
        }
        if (c == FinancialSeries.$$p[4]) {
            return this.kr(a, b, this.zh);
        }
        if (c == FinancialSeries.$$p[2]) {
            return this.kr(a, b, this.zd);
        }
        if (c == FinancialSeries.$$p[6]) {
            return this.kr(a, b, this.zl);
        }
        return _super.prototype.kq.call(this, a, b);
    };
    FinancialSeries.prototype.md = function (a) {
        var b = this.nd(a);
        if (b == FinancialSeries.$$p[11]) {
            return this.zw;
        }
        if (b == FinancialSeries.$$p[1]) {
            return this.zn;
        }
        if (b == FinancialSeries.$$p[9]) {
            return this.zs;
        }
        if (b == FinancialSeries.$$p[0]) {
            return this.za;
        }
        if (b == FinancialSeries.$$p[13]) {
            return this.z0;
        }
        if (b == FinancialSeries.$$p[5]) {
            return this.zj;
        }
        if (b == FinancialSeries.$$p[3]) {
            return this.zf;
        }
        if (b == FinancialSeries.$$p[4]) {
            return this.zh;
        }
        if (b == FinancialSeries.$$p[2]) {
            return this.zd;
        }
        if (b == FinancialSeries.$$p[6]) {
            return this.zl;
        }
        return _super.prototype.md.call(this, a);
    };
    FinancialSeries.prototype.y6 = function () {
        return 2;
    };
    FinancialSeries.prototype.y4 = function () {
        return 4;
    };
    FinancialSeries.prototype.aj = function (a) {
        var b = this.an(this.w6, this.w8.f, this.y3(this.cv), this.y5(this.cv), this.wh(a), true);
        if (b == null) {
            return null;
        }
        var c = b[0];
        var d = b[1];
        var e = new Array(1);
        var f = new Array(2);
        var g = this.y6();
        f[0] = { $type: Point_$type, x: c[0], y: c[g] };
        f[1] = { $type: Point_$type, x: d[0], y: d[g] };
        e[0] = f;
        return e;
    };
    FinancialSeries.prototype.al = function (a) {
        var b = this.an(this.w6, this.w8.f, this.y3(this.cv), this.y5(this.cv), this.wh(a), true);
        if (b == null) {
            return null;
        }
        var c = b[0];
        var d = b[1];
        var e = new Array(1);
        var f = new Array(2);
        f[0] = { $type: Point_$type, x: c[0], y: c[1] };
        f[1] = { $type: Point_$type, x: d[0], y: d[1] };
        e[0] = f;
        return e;
    };
    FinancialSeries.prototype.vw = function () {
        return this.aaw;
    };
    FinancialSeries.prototype.vx = function () {
        return this.aaw;
    };
    FinancialSeries.prototype.o2 = function () {
        _super.prototype.o2.call(this);
        this.w6 = null;
        this.xp = null;
    };
    FinancialSeries.prototype.get_ek = function () {
        return true;
    };
    FinancialSeries.prototype.get_f5 = function () {
        return this.xp != null && this.xp.ch;
    };
    Object.defineProperty(FinancialSeries.prototype, "x2", {
        get: function () {
            return this.xx;
        },
        set: function (a) {
            if (this.xx != a) {
                var b = this.xx;
                this.xx = a;
                this.raisePropertyChanged("OpenColumn", b, this.xx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeries.prototype, "x0", {
        get: function () {
            return this.xv;
        },
        set: function (a) {
            if (this.xv != a) {
                var b = this.xv;
                this.xv = a;
                this.raisePropertyChanged("HighColumn", b, this.xv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeries.prototype, "x1", {
        get: function () {
            return this.xw;
        },
        set: function (a) {
            if (this.xw != a) {
                var b = this.xw;
                this.xw = a;
                this.raisePropertyChanged("LowColumn", b, this.xw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeries.prototype, "xz", {
        get: function () {
            return this.xu;
        },
        set: function (a) {
            if (this.xu != a) {
                var b = this.xu;
                this.xu = a;
                this.raisePropertyChanged("CloseColumn", b, this.xu);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialSeries.prototype, "x4", {
        get: function () {
            return this.xy;
        },
        set: function (a) {
            if (this.xy != a) {
                var b = this.xy;
                this.xy = a;
                this.raisePropertyChanged("VolumeColumn", b, this.xy);
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialSeries.prototype.gb = function () {
        return this.highlightedItemsSource != null || (this.zj != null && this.zf != null && this.zh != null && this.zd != null);
    };
    FinancialSeries.prototype.ye = function () {
        return (this.assigningCategoryStyle != null && this.x7) || this.bx != 1 || this.cp != 1 || this.co != 1;
    };
    FinancialSeries.prototype.aad = function (a) {
        if (this.assigningCategoryStyle != null && this.x7) {
            this.assigningCategoryStyle(this, a);
        }
    };
    FinancialSeries.prototype.aab = function (a, b, c, d, e, f, g) {
        var h = this.bx != 1;
        var i = this.cp != 1;
        var j = this.co != 1;
        this.xf.a3(this.xf.b, d.cn, a, b, c, d, e, f, false, g);
        this.xf.ab(this.ib);
        var k = this.xf.ay;
        var l = null;
        var m = null;
        var n = null;
        var o = this.dn;
        if (h && this.db != null) {
            l = this.db.dl.getHighlightingInfo(this, o, k.z, k.y, false);
            k.c = l;
        }
        if (i && this.db != null) {
            m = this.db.dm.getHighlightingInfo(this, o, k.z, k.y, false);
            if (m == null && this.bn != null) {
                m = this.db.dm.getHighlightingInfo(this.bn, this.bn.dn, k.z, k.y, false);
            }
            k.d = m;
        }
        if (j && this.db != null) {
            n = this.db.dk.getHighlightingInfo(this, o, k.z, k.y, false);
            k.b = n;
        }
        k.h = f;
        this.aad(k);
        this.xf.z(this, l, m, n, f, h, i, j, b < 0);
    };
    FinancialSeries.prototype.yw = function (a) {
        return this.it(a, this.w6);
    };
    FinancialSeries.prototype.ka = function (a, b) {
        return this.kb(a, b, this.w6, runOn(this, this.yw), new FinancialValueList(this.x2, this.x0, this.x1, this.xz, this.x4));
    };
    FinancialSeries.prototype.j8 = function (a, b) {
        return this.j9(a, b, this.w6, runOn(this, this.yw), new FinancialValueList(this.x2, this.x0, this.x1, this.xz, this.x4));
    };
    FinancialSeries.prototype.ip = function (a, b, c, d, e) {
        if (c == null) {
            return Number.POSITIVE_INFINITY;
        }
        var f = this.w6.lj;
        return this.iq(a, b, this.w6, d, e, f, runOn(this, this.yw));
    };
    FinancialSeries.prototype.getOffsetValue = function () {
        return this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
    };
    FinancialSeries.prototype.getCategoryWidth = function () {
        return this.w6.getCategorySize(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
    };
    FinancialSeries.prototype.wc = function (a, b, c) {
        return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, null, null, null);
    };
    FinancialSeries.prototype.i5 = function (a, b, c) {
        var d = this.getEffectiveViewport1(this.cv);
        var e = this.yx(this.cv.by, this.cv.bx, d);
        var f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
        return this.i7(new FinancialValueList(this.x2, this.x0, this.x1, this.xz, this.x4), a, this.w6, f, e, runOn(this, this.yw), b, c);
    };
    FinancialSeries.prototype.i1 = function (a, b, c) {
        var d = this.getEffectiveViewport1(this.cv);
        var e = this.yx(this.cv.by, this.cv.bx, d);
        var f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
        return this.i7(this.x0, a, this.w6, f, e, runOn(this, this.yw), b, c);
    };
    FinancialSeries.prototype.i3 = function (a, b, c) {
        var d = this.getEffectiveViewport1(this.cv);
        var e = this.yx(this.cv.by, this.cv.bx, d);
        var f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
        return this.i7(this.x1, a, this.w6, f, e, runOn(this, this.yw), b, c);
    };
    FinancialSeries.prototype.yy = function (a, b, c) {
        var d = this.getEffectiveViewport1(this.cv);
        var e = this.yx(this.cv.by, this.cv.bx, d);
        var f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
        return this.i7(this.xz, a, this.w6, f, e, runOn(this, this.yw), b, c);
    };
    FinancialSeries.prototype.yz = function (a, b, c) {
        var d = this.getEffectiveViewport1(this.cv);
        var e = this.yx(this.cv.by, this.cv.bx, d);
        var f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
        return this.i7(this.x2, a, this.w6, f, e, runOn(this, this.yw), b, c);
    };
    FinancialSeries.prototype.y0 = function (a, b, c) {
        var d = this.getEffectiveViewport1(this.cv);
        var e = this.yx(this.cv.by, this.cv.bx, d);
        var f = new ScalerParams(0, this.cv.by, this.cv.bx, this.w6.ch, d);
        return this.i7(this.x4, a, this.w6, f, e, runOn(this, this.yw), b, c);
    };
    FinancialSeries.prototype.v8 = function (a, b, c) {
        var _this = this;
        return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, runOn(this, this.i1), function (d, e) { return _this.kb(d, e, _this.w6, runOn(_this, _this.yw), _this.x0); }, function (d, e) { return _this.j9(d, e, _this.w6, runOn(_this, _this.yw), _this.x0); });
    };
    FinancialSeries.prototype.wa = function (a, b, c) {
        var _this = this;
        return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, runOn(this, this.i3), function (d, e) { return _this.kb(d, e, _this.w6, runOn(_this, _this.yw), _this.x1); }, function (d, e) { return _this.j9(d, e, _this.w6, runOn(_this, _this.yw), _this.x1); });
    };
    FinancialSeries.prototype.aay = function (a, b, c) {
        var _this = this;
        return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, runOn(this, this.yz), function (d, e) { return _this.kb(d, e, _this.w6, runOn(_this, _this.yw), _this.x2); }, function (d, e) { return _this.j9(d, e, _this.w6, runOn(_this, _this.yw), _this.x2); });
    };
    FinancialSeries.prototype.aax = function (a, b, c) {
        var _this = this;
        return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, runOn(this, this.yy), function (d, e) { return _this.kb(d, e, _this.w6, runOn(_this, _this.yw), _this.xz); }, function (d, e) { return _this.j9(d, e, _this.w6, runOn(_this, _this.yw), _this.xz); });
    };
    FinancialSeries.prototype.aaz = function (a, b, c) {
        var _this = this;
        return this.we(a, b, c, this.yx(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.xp, this.w6, runOn(this, this.y0), function (d, e) { return _this.kb(d, e, _this.w6, runOn(_this, _this.yw), _this.x4); }, function (d, e) { return _this.j9(d, e, _this.w6, runOn(_this, _this.yw), _this.x4); });
    };
    FinancialSeries.prototype.qb = function () {
        _super.prototype.qb.call(this);
        if (this.w6 != null) {
            this.w6.ib(false);
        }
        if (this.xp != null) {
            this.xp.ib(false);
        }
    };
    FinancialSeries.prototype.sr = function (a, b) {
        this.xn.b0.e(this.id);
        this.rr(false);
    };
    FinancialSeries.prototype.sq = function (a, b) {
        if (this.g1 && this.jq < 0.05) {
            this.g2 = true;
            this.g5 = false;
            this.gk = false;
        }
        this.xn.b0.e(this.id);
        this.rr(false);
    };
    FinancialSeries.prototype.x3 = function (a, b, c) {
        var d = this.dk(b);
        this.yh.addItem(d, b);
        this.yi.addItem(b, c);
        return d;
    };
    FinancialSeries.prototype.z5 = function (a, b) {
        if (b == null) {
            return;
        }
        a.deregisterColumn(b);
        var c = this.yh.item(b);
        this.yi.removeItem(c);
        this.yh.removeItem(b);
    };
    FinancialSeries.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case "SyncLink":
                if (this.c3 != null && this.db != null) {
                    this.xn.b0.e(this.id);
                    this.rr(false);
                }
                break;
            case "SeriesViewer":
                if (c != null && d == null) {
                    if (this.w6 != null) {
                        this.w6.b0(this);
                    }
                    if (this.xp != null) {
                        this.xp.b0(this);
                    }
                }
                if (c == null && d != null) {
                    if (this.w6 != null) {
                        this.w6.ct(this);
                    }
                    if (this.xp != null) {
                        this.xp.ct(this);
                    }
                }
                this.xn.b0.e(this.id);
                this.rr(false);
                break;
            case "FastItemsSource":
                this.x6 = true;
                if (typeCast(IFastItemsSource_$type, c) != null) {
                    this.z5(typeCast(IFastItemsSource_$type, c), this.x2);
                    this.z5(typeCast(IFastItemsSource_$type, c), this.x0);
                    this.z5(typeCast(IFastItemsSource_$type, c), this.x1);
                    this.z5(typeCast(IFastItemsSource_$type, c), this.xz);
                    this.z5(typeCast(IFastItemsSource_$type, c), this.x4);
                    this.x2 = null;
                    this.x0 = null;
                    this.x1 = null;
                    this.xz = null;
                    this.x4 = null;
                }
                if (typeCast(IFastItemsSource_$type, d) != null) {
                    if (stringIsNullOrEmpty(this.zw) == false) {
                        this.x2 = this.x3(typeCast(IFastItemsSource_$type, d), this.zw, "OpenColumn");
                    }
                    if (stringIsNullOrEmpty(this.zn) == false) {
                        this.x0 = this.x3(typeCast(IFastItemsSource_$type, d), this.zn, "HighColumn");
                    }
                    if (stringIsNullOrEmpty(this.zs) == false) {
                        this.x1 = this.x3(typeCast(IFastItemsSource_$type, d), this.zs, "LowColumn");
                    }
                    if (stringIsNullOrEmpty(this.za) == false) {
                        this.xz = this.x3(typeCast(IFastItemsSource_$type, d), this.za, "CloseColumn");
                    }
                    if (stringIsNullOrEmpty(this.z0) == false) {
                        this.x4 = this.x3(typeCast(IFastItemsSource_$type, d), this.z0, "VolumeColumn");
                    }
                }
                this.x6 = false;
                if (this.xp != null && !this.xp.c1()) {
                    this.xn.b0.e(this.id);
                    this.rr(this.gg && c != null);
                }
                break;
            case FinancialSeries.$$p[11]:
                if (this.dn != null) {
                    this.z5(this.dn, this.x2);
                    this.x2 = this.x3(this.dn, this.zw, "OpenColumn");
                }
                break;
            case "OpenColumn":
                if (this.xp != null && !this.xp.c1() && !this.x6) {
                    this.xn.b0.e(this.id);
                    this.rr(this.gg && this.d7);
                }
                break;
            case FinancialSeries.$$p[1]:
                if (this.dn != null) {
                    this.z5(this.dn, this.x0);
                    this.x0 = this.x3(this.dn, this.zn, "HighColumn");
                }
                break;
            case "HighColumn":
                if (this.xp != null && !this.xp.c1() && !this.x6) {
                    this.xn.b0.e(this.id);
                    this.rr(this.gg && this.d7);
                }
                break;
            case FinancialSeries.$$p[9]:
                if (this.dn != null) {
                    this.z5(this.dn, this.x1);
                    this.x1 = this.x3(this.dn, this.zs, "LowColumn");
                }
                break;
            case "LowColumn":
                if (this.xp != null && !this.xp.c1() && !this.x6) {
                    this.xn.b0.e(this.id);
                    this.rr(this.gg && this.d7);
                }
                break;
            case FinancialSeries.$$p[0]:
                if (this.dn != null) {
                    this.z5(this.dn, this.xz);
                    this.xz = this.x3(this.dn, this.za, "CloseColumn");
                }
                break;
            case "CloseColumn":
                if (this.xp != null && !this.xp.c1() && !this.x6) {
                    this.xn.b0.e(this.id);
                    this.rr(this.gg && this.d7);
                }
                this.aaf();
                break;
            case FinancialSeries.$$p[13]:
                if (this.dn != null) {
                    this.z5(this.dn, this.x4);
                    this.x4 = this.x3(this.dn, this.z0, "VolumeColumn");
                }
                break;
            case "VolumeColumn":
                if (this.xp != null && !this.xp.c1() && !this.x6) {
                    this.xn.b0.e(this.id);
                    this.rr(this.gg && this.d7);
                }
                break;
            case FinancialSeries.$$p[14]:
                if (c != null) {
                    c.b0(this);
                }
                if (d != null) {
                    d.ct(this);
                }
                this.xn.b0.e(this.id);
                this.rr(false);
                this.ql();
                this.xo = typeCast(ITimeRangeSelectorAxis_$type, this.w6);
                break;
            case FinancialSeries.$$p[15]:
                if (c != null) {
                    c.b0(this);
                }
                if (d != null) {
                    d.ct(this);
                }
                this.xn.b0.e(this.id);
                if (this.xp != null) {
                    this.xp.c1();
                }
                {
                    this.rr(false);
                }
                this.ql();
                break;
            case "TransitionProgress":
                this.xc.c(this.jq, this.xa, this.w8);
                if (this.d5(this.cv)) {
                    return;
                }
                if (this.jq == 1) {
                    this.w8.w();
                    this.aae(this.w8, this.xn);
                    if (this.g1) {
                        this.g1 = false;
                        this.o6();
                        this.dc.r = truncate(this.i9());
                        this.dc.b = this.dd;
                    }
                    if (this.g4) {
                        this.g4 = false;
                        this.o6();
                        this.dc.r = truncate(this.i9());
                        this.dc.b = this.dd;
                        this.q0();
                    }
                    this.qo();
                }
                else {
                    this.aae(this.xc, this.xn);
                }
                if (this.db != null) {
                    this.db.qo(this);
                }
                break;
            case FinancialSeries.$$p[10]:
                this.rr(false);
                break;
            case "IsCustomCategoryStyleAllowed":
                this.x7 = this.ya();
                this.rr(false);
                break;
            case "CategoryMode":
                if (this.w6 != null) {
                    this.w6.ma(true);
                }
                break;
        }
    };
    FinancialSeries.prototype.x9 = function (a) {
        if (typeCast(NumericYAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    FinancialSeries.prototype.x8 = function (a) {
        if (typeCast(CategoryXAxis.$, a) !== null || a.ce) {
            return true;
        }
        return false;
    };
    FinancialSeries.prototype.wz = function (a, b, c) {
        if (b < a || a < 0 || a > this.dn.count || b < 0 || b > this.dn.count || c == null) {
            return null;
        }
        var d = new Array((b - a) + 1);
        var e = c.cn;
        var f = null;
        if (e) {
            f = c.i$b;
        }
        for (var g = a; g <= b; g++) {
            var h = g;
            if (e) {
                h = f._inner[h];
            }
            d[g - a] = this.dn.item(g);
        }
        return d;
    };
    FinancialSeries.prototype.y3 = function (a) {
        return a.b0.b;
    };
    FinancialSeries.prototype.y5 = function (a) {
        return a.b0.c;
    };
    FinancialSeries.prototype.wy = function (a, b) {
        return this.wz(a, b, this.xp);
    };
    FinancialSeries.prototype.get_fe = function () {
        return true;
    };
    FinancialSeries.prototype.o3 = function () {
        _super.prototype.o3.call(this);
        this.xt = null;
    };
    FinancialSeries.prototype.dg = function (a, b, c) {
        if (this.ed && this.fz(this.bx) && !c) {
            var d = this.xs(a, b, this.xt);
            this.xt = d;
            return d;
        }
        else {
            var e = this.xr(a, b, this.xt);
            this.xt = e;
            return e;
        }
    };
    FinancialSeries.prototype.y7 = function (a) {
        var b = this.w9();
        var c = this.cv.bx;
        var d = this.cv.by;
        var e = this.w6;
        var f = 0;
        if (e.cp) {
            var g = (a.y - d.top) / d.height;
            f = g * c.height + c.top;
        }
        else {
            var h = (a.x - d.left) / d.width;
            f = h * c.width + c.left;
        }
        var i = b.f.count;
        var j = 0;
        for (j = 0; j < i; j++) {
            var k = b.f._inner[j];
            var l = null;
            if (j < i - 1) {
                l = b.f._inner[j + 1];
            }
            if (k[0] <= f && l == null || l[0] >= f) {
                if (l != null) {
                    if (Math.abs(k[0] - f) < Math.abs(l[0] - f)) {
                        return j;
                    }
                    else {
                        return j + 1;
                    }
                }
                else {
                    return j;
                }
            }
        }
        return j;
    };
    FinancialSeries.prototype.xs = function (a, b, c) {
        if (this.dn == null) {
            return null;
        }
        var d = this.dn.count;
        var e = this.xn.b0.c;
        var f = this.xn.b0.d;
        var g = this.xn.b0.b;
        var h = this.w6;
        var i = -1;
        var j = -1;
        if (h.cn) {
            var k = this.w9();
            var l = this.y7(b);
            if (k.m != null && l >= 0 && l < k.m.count) {
                l = truncate(Math.round(k.m._inner[l]));
            }
            i = l;
            j = l;
        }
        else {
            var m = this.j3(b);
            i = truncate(Math.floor((intDivide(m, g)))) * g;
            j = i + (g - 1);
        }
        if (c != null && c.g == i && c.f == j) {
            return c;
        }
        var n = new HighlightingInfo();
        n.i = this;
        n.g = i;
        n.f = j;
        return n;
    };
    FinancialSeries.prototype.xr = function (a, b, c) {
        var d = new HighlightingInfo();
        d.i = this;
        d.g = 0;
        d.f = this.dn.count - 1;
        d.c = true;
        if (c != null && c.g == d.g && c.f == d.f) {
            return c;
        }
        return d;
    };
    FinancialSeries.prototype.ya = function () {
        return this.yc;
    };
    FinancialSeries.prototype.ir = function (a) {
        var b = this.cv.by;
        var c = this.cv.bx;
        var d = -1;
        if (!b.isEmpty && !c.isEmpty && this.w6 != null) {
            var e = this.getEffectiveViewport1(this.cv);
            var f = new ScalerParams(0, b, c, this.w6.ch, e);
            var g = this.wj(this.cv);
            var h = this.w6.eo(g.left, f);
            var i = this.w6.eo(g.right, f);
            var j = (a.x - b.left) / b.width;
            var k = h + (j * (i - h));
            if (this.w6.categoryMode != 0) {
                k -= 0.5;
            }
            d = k;
        }
        return d;
    };
    FinancialSeries.prototype.j3 = function (a) {
        return truncate(Math.round(this.ir(a)));
    };
    FinancialSeries.prototype.get_fs = function () {
        return typeCast(ISortingAxis_$type, this.w6) !== null;
    };
    FinancialSeries.prototype.a0 = function () {
        return this.w6;
    };
    FinancialSeries.prototype.j5 = function (a) {
        var b = this.cv.by;
        var c = this.cv.bx;
        var d = this.getEffectiveViewport1(this.cv);
        if (b.isEmpty || c.isEmpty) {
            return -1;
        }
        var e = new Rect(0, 0, 0, 1, 1);
        var f = new ScalerParams(0, b, c, this.w6.ch, d);
        var g = typeCast(ISortingAxis_$type, this.w6);
        var h = this.wj(this.cv);
        var i = h.left;
        var j = h.right;
        var k = this.v7(a).x;
        var l = this.w6.eo(k, f);
        var m = this.w6.eo(i, f);
        var n = this.w6.eo(j, f);
        l = Math.max(m, Math.min(n, l));
        var o = g.i$f(l);
        return o;
    };
    FinancialSeries.prototype.kn = function (a) {
        var b = 0;
        if (typeCast(ISortingAxis_$type, this.w6) !== null) {
            b = this.j5(a);
            if (b == -1) {
                return null;
            }
        }
        else {
            b = this.j3(a);
        }
        return b >= 0 && this.dn != null && b < this.dn.count ? this.dn.item(b) : null;
    };
    FinancialSeries.prototype.w9 = function () {
        var a = this.w8;
        if (this.d3()) {
            if (this.dc.g) {
                this.dc.t();
            }
            a = this.xc;
        }
        return a;
    };
    FinancialSeries.prototype.yx = function (a, b, c) {
        if (this.w6 == null) {
            return 0;
        }
        return this.w6.k9(a, b, c, this.index);
    };
    FinancialSeries.prototype.gn = function () {
        return this.yd;
    };
    FinancialSeries.prototype.xg = function () {
        return 11;
    };
    FinancialSeries.prototype.c4 = function () {
        return 3;
    };
    FinancialSeries.prototype.c7 = function () {
        return 3;
    };
    FinancialSeries.prototype.aae = function (a, b) {
        this.qx();
        this.wi = a.ab;
        b.aw();
    };
    FinancialSeries.prototype.aac = function (a, b) {
        a.v();
    };
    FinancialSeries.prototype.he = function (a, b, c) {
        var d = _super.prototype.he.call(this, a, b, c);
        var e = c;
        if (this.dn == null || this.dn.count == 0 || !c.o() || b.isEmpty || a.isEmpty || this.w6 == null || this.xp == null || e.b0.b < 1 || this.w6.bi == null || this.xp.bi == null || this.xp.l4 == this.xp.l3 || this.w6.itemsSource == null || this.w6.lj < 1) {
            e.b0.b = 0;
            d = false;
        }
        return d;
    };
    FinancialSeries.prototype.getViewInfo = function (a, b) {
        b = this.cv.by;
        a = this.cv.bx;
        return {
            p0: a,
            p1: b
        };
    };
    FinancialSeries.prototype.y1 = function (a) {
        var b = 0;
        var c = a.by;
        var d = a.bx;
        var e = this.getEffectiveViewport1(this.cv);
        if (!c.isEmpty && !d.isEmpty && this.xp != null) {
            var f = new ScalerParams(0, c, d, this.xp.ch, e);
            f.b = this.iy();
            b = this.xp.el(this.xp.me, f);
        }
        return b;
    };
    FinancialSeries.prototype.o6 = function () {
        _super.prototype.o6.call(this);
        this.xa.w();
        this.w8.w();
        this.xc.w();
    };
    FinancialSeries.prototype.rs = function (a) {
        var _this = this;
        _super.prototype.rs.call(this, a);
        this.xn.b0.e(this.id);
        if (this.d5(this.cv)) {
            return;
        }
        if (this.gr) {
            if (this.d3()) {
                this.aae(this.xc, this.xn);
            }
            else {
                this.aae(this.w8, this.xn);
            }
            return;
        }
        if (this.gf(a)) {
            var b = this.xa;
            if (this.d3()) {
                if (this.dc.g) {
                    this.dc.t();
                }
                this.xa = this.xc;
                this.xc = b;
            }
            else {
                this.xa = this.w8;
                this.w8 = b;
            }
            this.aac(this.w8, this.xn);
            var c = false;
            if (this.g3) {
                this.dc.w();
                this.dc.r = truncate(this.ja());
                this.dc.b = this.de != null ? this.de : this.dd;
                if (this.g4) {
                    var d = this.xc;
                    this.xc = this.xa;
                    this.xa = d;
                    c = true;
                }
                this.xi.s(this, this.xa, this.w8, this.isVertical, this.w6, this.xp, this.xh, this.xg(), this.c5, this.c4(), function () { return _this.y1(_this.xn); }, this.xn.bx);
                if (this.g4) {
                    var e = this.xc;
                    this.xc = this.xa;
                    this.xa = e;
                }
                this.g4 = false;
            }
            if (this.g6) {
                this.dc.w();
                this.dc.r = truncate(this.jb());
                this.dc.b = this.df != null ? this.df : this.dd;
                if (this.g1) {
                    var f = this.xc;
                    this.xc = this.xa;
                    this.xa = f;
                    c = true;
                }
                this.xi.s(this, this.xa, this.w8, this.isVertical, this.w6, this.xp, this.xh, this.xg(), this.c5, this.c4(), function () { return _this.y1(_this.xn); }, this.xn.bx);
                var g = this.w8;
                this.w8 = this.xa;
                this.xa = g;
                if (this.g1) {
                    var h = this.xc;
                    this.xc = this.xa;
                    this.xa = h;
                }
                this.g1 = false;
            }
            this.d9 = false;
            if (!c) {
                this.o1();
            }
            this.r7();
            if (this.g3) {
                this.g2 = false;
                this.g5 = false;
                this.gk = false;
                this.g1 = true;
            }
            if (this.g6) {
                this.g5 = false;
                this.g4 = true;
            }
        }
        else {
            this.aac(this.w8, this.xn);
            this.aae(this.w8, this.xn);
        }
    };
    FinancialSeries.prototype.go = function () {
        return true;
    };
    FinancialSeries.prototype.yf = function (a) {
        var e_1, _a;
        try {
            for (var _b = __values(fromEnum(a)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var b = _c.value;
                switch (b) {
                    case "HighColumn":
                        if (this.x0 == null) {
                            return false;
                        }
                        break;
                    case "LowColumn":
                        if (this.x1 == null) {
                            return false;
                        }
                        break;
                    case "OpenColumn":
                        if (this.x2 == null) {
                            return false;
                        }
                        break;
                    case "CloseColumn":
                        if (this.xz == null) {
                            return false;
                        }
                        break;
                    case "VolumeColumn":
                        if (this.x4 == null) {
                            return false;
                        }
                        break;
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        return true;
    };
    Object.defineProperty(FinancialSeries.prototype, "yg", {
        get: function () {
            return this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null;
        },
        enumerable: false,
        configurable: true
    });
    FinancialSeries.prototype.yb = function () {
        return this.typical != null;
    };
    FinancialSeries.prototype.z4 = function (a) {
        this.typical(this, a);
    };
    Object.defineProperty(FinancialSeries.prototype, "yo", {
        get: function () {
            return this.yl();
        },
        enumerable: false,
        configurable: true
    });
    FinancialSeries.prototype._yl = function () {
        var a, b, _a, _b, c, e_2_1, d, e, f, g, h;
        var e_2, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    a = false;
                    if (!this.yb())
                        return [3 /*break*/, 8];
                    b = this.xj(0, this.dn.count);
                    this.z4(new FinancialEventArgs(0, this.dn.count, b, this.xk(b), this));
                    if (!(b.typicalColumn != null))
                        return [3 /*break*/, 8];
                    a = true;
                    _d.label = 1;
                case 1:
                    _d.trys.push([1, 6, 7, 8]);
                    _a = __values(fromEnum(b.typicalColumn)), _b = _a.next();
                    _d.label = 2;
                case 2:
                    if (!!_b.done)
                        return [3 /*break*/, 5];
                    c = _b.value;
                    return [4 /*yield*/, c];
                case 3:
                    _d.sent();
                    _d.label = 4;
                case 4:
                    _b = _a.next();
                    return [3 /*break*/, 2];
                case 5: return [3 /*break*/, 8];
                case 6:
                    e_2_1 = _d.sent();
                    e_2 = { error: e_2_1 };
                    return [3 /*break*/, 8];
                case 7:
                    try {
                        if (_b && !_b.done && (_c = _a.return))
                            _c.call(_a);
                    }
                    finally {
                        if (e_2)
                            throw e_2.error;
                    }
                    return [7 /*endfinally*/];
                case 8:
                    if (!!a)
                        return [3 /*break*/, 17];
                    if (!(this.yg && this.w6.i$b != null))
                        return [3 /*break*/, 13];
                    d = this.dn.count;
                    e = this.w6.i$b;
                    f = 0;
                    _d.label = 9;
                case 9:
                    if (!(f < d))
                        return [3 /*break*/, 12];
                    return [4 /*yield*/, (this.x0.item(e.item(f)) + this.x1.item(e.item(f)) + this.xz.item(e.item(f))) / 3];
                case 10:
                    _d.sent();
                    _d.label = 11;
                case 11:
                    ++f;
                    return [3 /*break*/, 9];
                case 12: return [3 /*break*/, 17];
                case 13:
                    g = this.dn.count;
                    h = 0;
                    _d.label = 14;
                case 14:
                    if (!(h < g))
                        return [3 /*break*/, 17];
                    return [4 /*yield*/, (this.x0.item(h) + this.x1.item(h) + this.xz.item(h)) / 3];
                case 15:
                    _d.sent();
                    _d.label = 16;
                case 16:
                    ++h;
                    return [3 /*break*/, 14];
                case 17: return [2 /*return*/];
            }
        });
    };
    FinancialSeries.prototype.yl = function () {
        var _this = this;
        return toEnum(function () { return _this._yl(); });
    };
    Object.defineProperty(FinancialSeries.prototype, "yn", {
        get: function () {
            return this.yk();
        },
        enumerable: false,
        configurable: true
    });
    FinancialSeries.prototype._yk = function () {
        var a, b, c, d, e;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!(this.yg && this.w6.i$b != null))
                        return [3 /*break*/, 7];
                    a = this.x0.count;
                    b = this.w6.i$b;
                    if (!(a > 0))
                        return [3 /*break*/, 2];
                    return [4 /*yield*/, this.y2(this.x0.item(b.item(0)) - this.x1.item(b.item(0)))];
                case 1:
                    _a.sent();
                    _a.label = 2;
                case 2:
                    c = 1;
                    _a.label = 3;
                case 3:
                    if (!(c < a))
                        return [3 /*break*/, 6];
                    return [4 /*yield*/, Math.max(this.y2(this.x0.item(b.item(c)) - this.x1.item(b.item(c))), Math.max(this.y2(Math.abs(this.x0.item(b.item(c)) - this.xz.item(b.item(c - 1)))), this.y2(Math.abs(this.x1.item(b.item(c)) - this.xz.item(b.item(c - 1))))))];
                case 4:
                    _a.sent();
                    _a.label = 5;
                case 5:
                    ++c;
                    return [3 /*break*/, 3];
                case 6: return [3 /*break*/, 13];
                case 7:
                    d = this.x0.count;
                    if (!(d > 0))
                        return [3 /*break*/, 9];
                    return [4 /*yield*/, this.y2(this.x0.item(0) - this.x1.item(0))];
                case 8:
                    _a.sent();
                    _a.label = 9;
                case 9:
                    e = 1;
                    _a.label = 10;
                case 10:
                    if (!(e < d))
                        return [3 /*break*/, 13];
                    return [4 /*yield*/, Math.max(this.y2(this.x0.item(e) - this.x1.item(e)), Math.max(this.y2(Math.abs(this.x0.item(e) - this.xz.item(e - 1))), this.y2(Math.abs(this.x1.item(e) - this.xz.item(e - 1)))))];
                case 11:
                    _a.sent();
                    _a.label = 12;
                case 12:
                    ++e;
                    return [3 /*break*/, 10];
                case 13: return [2 /*return*/];
            }
        });
    };
    FinancialSeries.prototype.yk = function () {
        var _this = this;
        return toEnum(function () { return _this._yk(); });
    };
    Object.defineProperty(FinancialSeries.prototype, "ym", {
        get: function () {
            return this.yj();
        },
        enumerable: false,
        configurable: true
    });
    FinancialSeries.prototype._yj = function () {
        var a, b, c, d, e;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!(this.yg && this.w6.i$b != null))
                        return [3 /*break*/, 7];
                    a = this.x1.count;
                    b = this.w6.i$b;
                    if (!(a > 0))
                        return [3 /*break*/, 2];
                    return [4 /*yield*/, this.y2(this.x1.item(b.item(0)))];
                case 1:
                    _a.sent();
                    _a.label = 2;
                case 2:
                    c = 1;
                    _a.label = 3;
                case 3:
                    if (!(c < a))
                        return [3 /*break*/, 6];
                    return [4 /*yield*/, Math.min(this.y2(this.x1.item(b.item(c))), this.y2(this.xz.item(b.item(c - 1))))];
                case 4:
                    _a.sent();
                    _a.label = 5;
                case 5:
                    c++;
                    return [3 /*break*/, 3];
                case 6: return [3 /*break*/, 13];
                case 7:
                    d = this.x1.count;
                    if (!(d > 0))
                        return [3 /*break*/, 9];
                    return [4 /*yield*/, this.y2(this.x1.item(0))];
                case 8:
                    _a.sent();
                    _a.label = 9;
                case 9:
                    e = 1;
                    _a.label = 10;
                case 10:
                    if (!(e < d))
                        return [3 /*break*/, 13];
                    return [4 /*yield*/, Math.min(this.y2(this.x1.item(e)), this.y2(this.xz.item(e - 1)))];
                case 11:
                    _a.sent();
                    _a.label = 12;
                case 12:
                    e++;
                    return [3 /*break*/, 10];
                case 13: return [2 /*return*/];
            }
        });
    };
    FinancialSeries.prototype.yj = function () {
        var _this = this;
        return toEnum(function () { return _this._yj(); });
    };
    FinancialSeries.prototype.yq = function (a) {
        if (a == null) {
            return null;
        }
        if (this.yg && this.w6.i$b != null) {
            return new SafeSortedReadOnlyDoubleCollection(0, a, this.w6.i$b);
        }
        return new SafeReadOnlyDoubleCollection(0, a);
    };
    FinancialSeries.prototype.yp = function () {
        var a = new List$1(String_$type, 0);
        a.add("HighColumn");
        a.add("LowColumn");
        a.add("CloseColumn");
        if (this.typicalBasedOn != null && this.dn != null && !this.x5) {
            this.x5 = true;
            var b = this.xj(0, this.dn.count);
            this.x5 = false;
            var c = new FinancialEventArgs(0, this.dn.count, b, this.xk(b), this);
            this.typicalBasedOn(this, c);
            if (c.basedOn != null && c.basedOn.count > 0) {
                return c.basedOn;
            }
        }
        return a;
    };
    FinancialSeries.prototype.getOpenColumnAsArray = function () {
        if (this.x2 == null) {
            return null;
        }
        if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
            if (this.yu != null) {
                return this.yu.asArray();
            }
            var a = new List$1(Number_$type, 2, this.x2.count);
            var b = this.w6.i$b;
            var c = this.x2.asArray();
            for (var d = 0; d < b.count; d++) {
                a.add(c[b._inner[d]]);
            }
            this.yu = a;
            return this.yu.asArray();
        }
        else {
            return this.x2.asArray();
        }
    };
    FinancialSeries.prototype.getHighColumnAsArray = function () {
        if (this.x0 == null) {
            return null;
        }
        if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
            if (this.ys != null) {
                return this.ys.asArray();
            }
            var a = new List$1(Number_$type, 2, this.x0.count);
            var b = this.w6.i$b;
            var c = this.x0.asArray();
            for (var d = 0; d < b.count; d++) {
                a.add(c[b._inner[d]]);
            }
            this.ys = a;
            return this.ys.asArray();
        }
        else {
            return this.x0.asArray();
        }
    };
    FinancialSeries.prototype.getLowColumnAsArray = function () {
        if (this.x1 == null) {
            return null;
        }
        if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
            if (this.yt != null) {
                return this.yt.asArray();
            }
            var a = new List$1(Number_$type, 2, this.x1.count);
            var b = this.w6.i$b;
            var c = this.x1.asArray();
            for (var d = 0; d < b.count; d++) {
                a.add(c[b._inner[d]]);
            }
            this.yt = a;
            return this.yt.asArray();
        }
        else {
            return this.x1.asArray();
        }
    };
    FinancialSeries.prototype.getCloseColumnAsArray = function () {
        if (this.xz == null) {
            return null;
        }
        if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
            if (this.yr != null) {
                return this.yr.asArray();
            }
            var a = new List$1(Number_$type, 2, this.xz.count);
            var b = this.w6.i$b;
            var c = this.xz.asArray();
            for (var d = 0; d < b.count; d++) {
                a.add(c[b._inner[d]]);
            }
            this.yr = a;
            return this.yr.asArray();
        }
        else {
            return this.xz.asArray();
        }
    };
    FinancialSeries.prototype.getVolumeColumnAsArray = function () {
        if (this.x4 == null) {
            return null;
        }
        if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
            if (this.yv != null) {
                return this.yv.asArray();
            }
            var a = new List$1(Number_$type, 2, this.x4.count);
            var b = this.w6.i$b;
            var c = this.x4.asArray();
            for (var d = 0; d < b.count; d++) {
                a.add(c[b._inner[d]]);
            }
            this.yv = a;
            return this.yv.asArray();
        }
        else {
            return this.x4.asArray();
        }
    };
    FinancialSeries.prototype.pb = function (a, b, c, d) {
        _super.prototype.pb.call(this, a, b, c, d);
        if (this.w6 == null || !(typeCast(ISortingAxis_$type, this.w6) !== null)) {
            return;
        }
        var e = this.w6.i$b;
        switch (a) {
            case 1:
                for (var f = b; f < c; f++) {
                    var g = e._inner[f];
                    if (this.yu != null) {
                        this.yu.insert(f, this.x2.item(g));
                    }
                    if (this.ys != null) {
                        this.ys.insert(f, this.x0.item(g));
                    }
                    if (this.yt != null) {
                        this.yt.insert(f, this.x1.item(g));
                    }
                    if (this.yr != null) {
                        this.yr.insert(f, this.xz.item(g));
                    }
                    if (this.yv != null) {
                        this.yv.insert(f, this.x4.item(g));
                    }
                }
                break;
            case 0:
                if (this.yu != null) {
                    this.yu.v(b, c);
                }
                if (this.ys != null) {
                    this.ys.v(b, c);
                }
                if (this.yt != null) {
                    this.yt.v(b, c);
                }
                if (this.yr != null) {
                    this.yr.v(b, c);
                }
                if (this.yv != null) {
                    this.yv.v(b, c);
                }
                break;
            case 4:
                if (this.yu != null) {
                    this.yu = new List$1(Number_$type, 0);
                }
                if (this.ys != null) {
                    this.ys = new List$1(Number_$type, 0);
                }
                if (this.yt != null) {
                    this.yt = new List$1(Number_$type, 0);
                }
                if (this.yr != null) {
                    this.yr = new List$1(Number_$type, 0);
                }
                if (this.yv != null) {
                    this.yv = new List$1(Number_$type, 0);
                }
                break;
            case 3:
            case 2:
                for (var h = b; h < c; h++) {
                    var i = e._inner[h];
                    if (this.yu != null) {
                        this.yu._inner[h] = this.x2.item(i);
                    }
                    if (this.ys != null) {
                        this.ys._inner[h] = this.x0.item(i);
                    }
                    if (this.yt != null) {
                        this.yt._inner[h] = this.x1.item(i);
                    }
                    if (this.yr != null) {
                        this.yr._inner[h] = this.xz.item(i);
                    }
                    if (this.yv != null) {
                        this.yv._inner[h] = this.x4.item(i);
                    }
                }
                break;
        }
        this.aaf();
        this.sn();
    };
    FinancialSeries.prototype.xj = function (a, b) {
        var _this = this;
        var c = this.yq(this.x2);
        var d = this.yq(this.xz);
        var e = this.yq(this.x0);
        var f = this.yq(this.x1);
        var g = this.yq(this.x4);
        var h = ((function () {
            var $ret = new FinancialCalculationDataSource();
            $ret.typicalColumn = new CalculatedColumn(1, new SafeEnumerable(_this.yo), _this.yp());
            $ret.trueRange = new CalculatedColumn(1, new SafeEnumerable(_this.yn), ((function () {
                var $ret = new List$1(String_$type, 0);
                $ret.add("HighColumn");
                $ret.add("LowColumn");
                $ret.add("CloseColumn");
                return $ret;
            })()));
            $ret.trueLow = new CalculatedColumn(1, new SafeEnumerable(_this.ym), ((function () {
                var $ret = new List$1(String_$type, 0);
                $ret.add("LowColumn");
                $ret.add("CloseColumn");
                return $ret;
            })()));
            $ret.openColumn = c;
            $ret.closeColumn = d;
            $ret.highColumn = e;
            $ret.lowColumn = f;
            $ret.volumeColumn = g;
            $ret.calculateFrom = a;
            $ret.calculateCount = b;
            $ret.minimumValue = NaN;
            $ret.maximumValue = NaN;
            $ret.count = _this.dn != null ? _this.dn.count : 0;
            return $ret;
        })());
        return h;
    };
    FinancialSeries.prototype.y2 = function (a) {
        if (isInfinity(a) || isNaN_(a)) {
            return 0;
        }
        return a;
    };
    FinancialSeries.prototype.xk = function (a) {
        var _this = this;
        return ((function () {
            var $ret = new FinancialCalculationSupportingCalculations();
            $ret.a = new ColumnSupportingCalculation(1, TrendCalculators.b, new List$1(String_$type, 0));
            $ret.c = new ColumnSupportingCalculation(1, TrendCalculators.e, new List$1(String_$type, 0));
            $ret.d = new ColumnSupportingCalculation(1, TrendCalculators.f, new List$1(String_$type, 0));
            $ret.b = new ColumnSupportingCalculation(1, TrendCalculators.d, new List$1(String_$type, 0));
            $ret.h = new DataSourceSupportingCalculation(1, function (b) { return TrendCalculators.b(b.volumeColumn, b.shortPeriod); }, ((function () {
                var $ret = new List$1(String_$type, 0);
                $ret.add("VolumeColumn");
                return $ret;
            })()));
            $ret.f = new DataSourceSupportingCalculation(1, function (b) { return TrendCalculators.b(b.volumeColumn, b.longPeriod); }, ((function () {
                var $ret = new List$1(String_$type, 0);
                $ret.add("VolumeColumn");
                return $ret;
            })()));
            $ret.g = new DataSourceSupportingCalculation(1, function (b) { return TrendCalculators.b(b.typicalColumn, b.shortPeriod); }, a.typicalColumn.basedOn);
            $ret.e = new DataSourceSupportingCalculation(1, function (b) { return TrendCalculators.b(b.typicalColumn, b.longPeriod); }, a.typicalColumn.basedOn);
            $ret.k = Series.hq;
            $ret.j = Series.hp;
            $ret.i = runOn(_this, _this.y2);
            return $ret;
        })());
    };
    Object.defineProperty(FinancialSeries.prototype, "categoryAxis", {
        get: function () {
            return this.w6;
        },
        enumerable: false,
        configurable: true
    });
    FinancialSeries.prototype.rt = function (a, b) {
        _super.prototype.rt.call(this, a, b);
        if (!this.g0) {
            this.cv.a4(b);
            return;
        }
        var c = typeCast(FinancialSeriesView.$, this.cu);
        this.cv.a4(b);
        c.b0.e(this.id);
        if (this.d5(this.cu)) {
            return;
        }
        if (!this.gs) {
            this.aac(this.xb, c);
        }
        this.gs = false;
        this.aae(this.xb, c);
        this.g0 = false;
    };
    FinancialSeries.prototype.rq = function (a, b, c, d, e) {
        _super.prototype.rq.call(this, a, b, c, d, e);
        var f = this.ba.a$j.item(d);
        var g = f;
        g.b0.e(this.id);
        f.a3(c);
        if (this.d5(g)) {
            return;
        }
        if (this.w7 == null) {
            this.w7 = new CategoryFrame(3);
        }
        this.w7.v();
        this.aac(this.w7, g);
        this.aae(this.w7, g);
    };
    FinancialSeries.prototype.aw = function (a) {
        if (this.w6 == a) {
            return this.xp;
        }
        if (this.xp == a) {
            return this.w6;
        }
        return null;
    };
    FinancialSeries.prototype.preferredCategoryMode = function (a) {
        var b = typeCast(IDateTimeAxis_$type, this.w6);
        return b != null ? b.preferredCategoryMode(this) : this.xe;
    };
    Object.defineProperty(FinancialSeries.prototype, "xe", {
        get: function () {
            return this.xd;
        },
        set: function (a) {
            var b = this.xe;
            if (a != b) {
                this.xd = a;
                this.raisePropertyChanged("CategoryMode", enumGetBox(CategoryMode_$type, b), enumGetBox(CategoryMode_$type, this.xe));
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialSeries.prototype.get_fm = function () {
        return true;
    };
    FinancialSeries.prototype.r3 = function (a, b) {
        _super.prototype.r3.call(this, a, b);
        this.aaw = a;
    };
    FinancialSeries.prototype.aaf = function () {
        this.im = this.xz == null || this.xz.count == 0 ? NaN : this.xz.item(this.xz.count - 1);
    };
    FinancialSeries.prototype.kp = function (a) {
        if (this.xo != null && a != null) {
            if (this.xq == null || this.xq.propertyName != this.xo.dateTimeMemberPath) {
                this.xq = new FastReflectionHelper(false, this.xo.dateTimeMemberPath);
            }
            var b = this.xq.getPropertyValue(a);
            if (typeCast(Date_$type, b) !== null) {
                return dateToStringFormat(b, this.xo.itemLabelFormat());
            }
        }
        return _super.prototype.kp.call(this, a);
    };
    FinancialSeries.$t = markType(FinancialSeries, 'FinancialSeries', Series.$, [IHasCategoryAxis_$type, IHasCategoryModePreference_$type]);
    FinancialSeries.$$p = markDep(DependencyProperty, PropertyMetadata, FinancialSeries, 'raisePropertyChanged', ['CloseMemberPath:za:aag', [2, null], 'HighMemberPath:zn:aam', [2, null], 'HighlightedCloseMemberPath:zd:aah', [2, null], 'HighlightedHighMemberPath:zf:aai', [2, null], 'HighlightedLowMemberPath:zh:aaj', [2, null], 'HighlightedOpenMemberPath:zj:aak', [2, null], 'HighlightedVolumeMemberPath:zl:aal', [2, null], 'IsCustomCategoryStyleAllowed:yc:aan', [0, false], 'IsTransitionInEnabled:yd:aao', [0, false], 'LowMemberPath:zs:aap', [2, null], 'NegativeBrush:aaw:aaq', [Brush.$, null], 'OpenMemberPath:zw:aar', [2, null], 'TransitionInMode:xh:aas', [CategoryTransitionInMode_$type, enumGetBox(CategoryTransitionInMode_$type, 0)], 'VolumeMemberPath:z0:aat', [2, null], 'XAxis:w6:aau', [CategoryAxisBase.$, null], 'YAxis:xp:aav', [NumericYAxis.$, null]]);
    return FinancialSeries;
}(Series));
export { FinancialSeries };
/**
 * @hidden
 */
var FinancialCalculationSupportingCalculations = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialCalculationSupportingCalculations, _super);
    function FinancialCalculationSupportingCalculations() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.a = null;
        _this.c = null;
        _this.d = null;
        _this.b = null;
        _this.h = null;
        _this.f = null;
        _this.g = null;
        _this.e = null;
        _this.k = null;
        _this.j = null;
        _this.i = null;
        return _this;
    }
    FinancialCalculationSupportingCalculations.$t = markType(FinancialCalculationSupportingCalculations, 'FinancialCalculationSupportingCalculations');
    return FinancialCalculationSupportingCalculations;
}(Base));
export { FinancialCalculationSupportingCalculations };
/**
 * @hidden
 */
var DataSourceSupportingCalculation = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataSourceSupportingCalculation, _super);
    function DataSourceSupportingCalculation(a) {
        var _rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            _rest[_i - 1] = arguments[_i];
        }
        var _this = this;
        a = (a == void 0) ? 0 : a;
        switch (a) {
            case 0:
                {
                    var c = _rest[0];
                    _this = _super.call(this, Delegate_$type, 0, c) || this;
                }
                break;
            case 1:
                {
                    var c = _rest[0];
                    var d = _rest[1];
                    _this = _super.call(this, Delegate_$type, 1, c, d) || this;
                }
                break;
        }
        return _this;
    }
    DataSourceSupportingCalculation.$t = markType(DataSourceSupportingCalculation, 'DataSourceSupportingCalculation', SupportingCalculation$1.$.specialize(Delegate_$type));
    return DataSourceSupportingCalculation;
}(SupportingCalculation$1));
export { DataSourceSupportingCalculation };
/**
 * @hidden
 */
var FinancialCalculationDataSource = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialCalculationDataSource, _super);
    function FinancialCalculationDataSource() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._openColumn = null;
        _this._closeColumn = null;
        _this._highColumn = null;
        _this._lowColumn = null;
        _this._volumeColumn = null;
        _this._indicatorColumn = null;
        _this._typicalColumn = null;
        _this._trueRange = null;
        _this._trueLow = null;
        _this._period = 0;
        _this._shortPeriod = 0;
        _this._longPeriod = 0;
        _this._count = 0;
        _this._calculateFrom = 0;
        _this._calculateCount = 0;
        _this._multiplier = 0;
        _this._minimumValue = 0;
        _this._maximumValue = 0;
        _this._specifiesRange = false;
        _this._series = null;
        return _this;
    }
    Object.defineProperty(FinancialCalculationDataSource.prototype, "openColumn", {
        get: function () {
            return this._openColumn;
        },
        set: function (a) {
            this._openColumn = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "closeColumn", {
        get: function () {
            return this._closeColumn;
        },
        set: function (a) {
            this._closeColumn = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "highColumn", {
        get: function () {
            return this._highColumn;
        },
        set: function (a) {
            this._highColumn = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "lowColumn", {
        get: function () {
            return this._lowColumn;
        },
        set: function (a) {
            this._lowColumn = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "volumeColumn", {
        get: function () {
            return this._volumeColumn;
        },
        set: function (a) {
            this._volumeColumn = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "indicatorColumn", {
        get: function () {
            return this._indicatorColumn;
        },
        set: function (a) {
            this._indicatorColumn = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "typicalColumn", {
        get: function () {
            return this._typicalColumn;
        },
        set: function (a) {
            this._typicalColumn = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "trueRange", {
        get: function () {
            return this._trueRange;
        },
        set: function (a) {
            this._trueRange = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "trueLow", {
        get: function () {
            return this._trueLow;
        },
        set: function (a) {
            this._trueLow = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "period", {
        get: function () {
            return this._period;
        },
        set: function (a) {
            this._period = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "shortPeriod", {
        get: function () {
            return this._shortPeriod;
        },
        set: function (a) {
            this._shortPeriod = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "longPeriod", {
        get: function () {
            return this._longPeriod;
        },
        set: function (a) {
            this._longPeriod = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "count", {
        get: function () {
            return this._count;
        },
        set: function (a) {
            this._count = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "calculateFrom", {
        get: function () {
            return this._calculateFrom;
        },
        set: function (a) {
            this._calculateFrom = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "calculateCount", {
        get: function () {
            return this._calculateCount;
        },
        set: function (a) {
            this._calculateCount = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "multiplier", {
        get: function () {
            return this._multiplier;
        },
        set: function (a) {
            this._multiplier = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "minimumValue", {
        get: function () {
            return this._minimumValue;
        },
        set: function (a) {
            this._minimumValue = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "maximumValue", {
        get: function () {
            return this._maximumValue;
        },
        set: function (a) {
            this._maximumValue = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "specifiesRange", {
        get: function () {
            return this._specifiesRange;
        },
        set: function (a) {
            this._specifiesRange = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialCalculationDataSource.prototype, "series", {
        get: function () {
            return this._series;
        },
        set: function (a) {
            this._series = a;
        },
        enumerable: false,
        configurable: true
    });
    FinancialCalculationDataSource.$t = markType(FinancialCalculationDataSource, 'FinancialCalculationDataSource');
    return FinancialCalculationDataSource;
}(Base));
export { FinancialCalculationDataSource };
/**
 * @hidden
 */
var FinancialSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialSeriesView, _super);
    function FinancialSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.b2 = null;
        _this.b0 = null;
        _this.b4 = 0;
        _this.b4 = -1;
        _this.b2 = a;
        _this.b0 = _this.b1();
        return _this;
    }
    FinancialSeriesView.prototype.b3 = function (a) {
        if (this.b4 != a.u) {
            return true;
        }
        return false;
    };
    FinancialSeriesView.prototype.b5 = function (a) {
        this.b4 = a.u;
    };
    FinancialSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
        }
    };
    FinancialSeriesView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'>";
        if (this.b2.w6.ce) {
            a += "<span>${itemLabel}</span><br/>";
        }
        else if (this.b2.w6.f2 != null) {
            a += "<span>${item." + this.b2.w6.f2 + "}</span><br/>";
        }
        a += "<span class='ui-priority-primary'";
        var b = this.e.ba.a$u(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        var c = new SRProvider(this.w);
        var d = c.b1("financialSeries_Open");
        var e = c.b1("financialSeries_High");
        var f = c.b1("financialSeries_Low");
        var g = c.b1("financialSeries_Close");
        d = stringIsNullOrEmpty(d) ? "Open" : d;
        e = stringIsNullOrEmpty(e) ? "High" : e;
        f = stringIsNullOrEmpty(f) ? "Low" : f;
        g = stringIsNullOrEmpty(g) ? "Close" : g;
        a += ">" + this.b2.title + "</span><table><tr><td>" + d + ":</td><td>" + "${item." + this.b2.zw + "}</td></tr><tr><td>" + e + ":</td><td>" + "${item." + this.b2.zn + "}</td></tr><tr><td>" + f + ":</td><td>" + "${item." + this.b2.zs + "}</td></tr><tr><td>" + g + ":</td><td>" + "${item." + this.b2.za + "}</td></tr>";
        if (!stringIsNullOrEmpty(this.b2.z0)) {
            var h = c.b1("financialSeries_Volume");
            h = stringIsNullOrEmpty(h) ? "Volume" : h;
            a += "<tr><td>" + h + ":</td><td>" + "${item." + this.b2.z0 + "}</td></tr>";
        }
        a += "</table></div>";
        return a;
    };
    FinancialSeriesView.prototype.af = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.v1 = a;
        this.e.jm = 5;
        this.e.jn = 2;
        this.e.jo = 2;
    };
    FinancialSeriesView.$t = markType(FinancialSeriesView, 'FinancialSeriesView', SeriesView.$);
    return FinancialSeriesView;
}(SeriesView));
export { FinancialSeriesView };
