// igniteui-react-charts/esm5/lib/FinancialChartDropDownControlView.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, markType } from "igniteui-react-core";
import { SRProvider } from "igniteui-react-core";
/**
 * @hidden
 */
var FinancialChartDropDownControlView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialChartDropDownControlView, _super);
    function FinancialChartDropDownControlView() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.b = false;
        _this.c = false;
        _this.d = null;
        _this.e = null;
        _this.a = null;
        return _this;
    }
    FinancialChartDropDownControlView.prototype.i = function (a) {
        if (!this.b && !this.c) {
            this.g();
        }
        this.b = false;
        this.c = false;
    };
    FinancialChartDropDownControlView.prototype.h = function (a) {
        this.b = true;
    };
    FinancialChartDropDownControlView.prototype.k = function (a) {
    };
    FinancialChartDropDownControlView.prototype.j = function (a) {
        this.c = true;
        this.k(a.target);
    };
    FinancialChartDropDownControlView.prototype.f = function (a, b, c, d) {
        this.a = b;
        this.a.a = new SRProvider(a);
    };
    FinancialChartDropDownControlView.prototype.g = function () {
        this.a.isOpen = false;
        this.a.updateUI();
    };
    FinancialChartDropDownControlView.$t = markType(FinancialChartDropDownControlView, 'FinancialChartDropDownControlView');
    return FinancialChartDropDownControlView;
}(Base));
export { FinancialChartDropDownControlView };
