// igniteui-react-charts/esm5/lib/FinalValueSelectionMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Gets or sets how to select the final value to annotate.
 */
export var FinalValueSelectionMode = /*@__PURE__*/ (function (FinalValueSelectionMode) {
    /**
     * A mode is selected automatically
     */
    FinalValueSelectionMode[FinalValueSelectionMode["Auto"] = 0] = "Auto";
    /**
     * Displays the last value visible.
     */
    FinalValueSelectionMode[FinalValueSelectionMode["FinalVisible"] = 1] = "FinalVisible";
    /**
     * Displays an interpolated last value for when the series leaves view.
     */
    FinalValueSelectionMode[FinalValueSelectionMode["FinalVisibleInterpolated"] = 2] = "FinalVisibleInterpolated";
    /**
     * Displays the last value in the series, whether visible or not.
     */
    FinalValueSelectionMode[FinalValueSelectionMode["Final"] = 3] = "Final";
    return FinalValueSelectionMode;
})({});
/**
 * @hidden
 */
export var FinalValueSelectionMode_$type = /*@__PURE__*/ markEnum('FinalValueSelectionMode', 'Auto,0|FinalVisible,1|FinalVisibleInterpolated,2|Final,3');
