// igniteui-react-charts/esm5/lib/FinancialChartRangeSelectorOption.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * An enum representing the time range presets on a financial chart range selector.
 */
export var FinancialChartRangeSelectorOption = /*@__PURE__*/ (function (FinancialChartRangeSelectorOption) {
    /**
     * Include a preset button for one month in the range selector.
     */
    FinancialChartRangeSelectorOption[FinancialChartRangeSelectorOption["OneMonth"] = 0] = "OneMonth";
    /**
     * Include a preset button for three months in the range selector.
     */
    FinancialChartRangeSelectorOption[FinancialChartRangeSelectorOption["ThreeMonths"] = 1] = "ThreeMonths";
    /**
     * Include a preset button for six months in the range selector.
     */
    FinancialChartRangeSelectorOption[FinancialChartRangeSelectorOption["SixMonths"] = 2] = "SixMonths";
    /**
     * Include a preset button for year-to-date in the range selector.
     */
    FinancialChartRangeSelectorOption[FinancialChartRangeSelectorOption["YearToDate"] = 3] = "YearToDate";
    /**
     * Include a preset button for one year in the range selector.
     */
    FinancialChartRangeSelectorOption[FinancialChartRangeSelectorOption["OneYear"] = 4] = "OneYear";
    /**
     * Include a preset button for all data in the range selector.
     */
    FinancialChartRangeSelectorOption[FinancialChartRangeSelectorOption["All"] = 5] = "All";
    return FinancialChartRangeSelectorOption;
})({});
/**
 * @hidden
 */
export var FinancialChartRangeSelectorOption_$type = /*@__PURE__*/ markEnum('FinancialChartRangeSelectorOption', 'OneMonth,0|ThreeMonths,1|SixMonths,2|YearToDate,3|OneYear,4|All,5');
