// igniteui-react-charts/esm5/lib/FinancialChartType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * An enum representing the available price series types to display in the main chart.
 */
export var FinancialChartType = /*@__PURE__*/ (function (FinancialChartType) {
    /**
     * Automatically determine the price series type to display in the main chart.
     */
    FinancialChartType[FinancialChartType["Auto"] = 0] = "Auto";
    /**
     * Display financial bar series in the main chart.
     */
    FinancialChartType[FinancialChartType["Bar"] = 1] = "Bar";
    /**
     * Display candlestick series in the main chart.
     */
    FinancialChartType[FinancialChartType["Candle"] = 2] = "Candle";
    /**
     * Display column series in the main chart.
     */
    FinancialChartType[FinancialChartType["Column"] = 3] = "Column";
    /**
     * Display line series in the main chart.
     */
    FinancialChartType[FinancialChartType["Line"] = 4] = "Line";
    return FinancialChartType;
})({});
/**
 * @hidden
 */
export var FinancialChartType_$type = /*@__PURE__*/ markEnum('FinancialChartType', 'Auto,0|Bar,1|Candle,2|Column,3|Line,4');
