// igniteui-react-charts/esm5/lib/ForceIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ForceIndexIndicatorStrategy } from "./ForceIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var ForceIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ForceIndexIndicator, _super);
    function ForceIndexIndicator() {
        return _super.call(this) || this;
    }
    ForceIndexIndicator.prototype.bj = function () {
        return new ForceIndexIndicator();
    };
    ForceIndexIndicator.prototype.get_aa2 = function () {
        return 1;
    };
    ForceIndexIndicator.prototype.get_abv = function () {
        return new ForceIndexIndicatorStrategy();
    };
    Object.defineProperty(ForceIndexIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    ForceIndexIndicator.prototype.get_ab8 = function () {
        return ForceIndexIndicator.$;
    };
    Object.defineProperty(ForceIndexIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ForceIndexIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(ForceIndexIndicator.ach));
        },
        set: function (a) {
            this.h(ForceIndexIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    ForceIndexIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    ForceIndexIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    ForceIndexIndicator.$t = markType(ForceIndexIndicator, 'ForceIndexIndicator', StrategyBasedIndicator.$);
    ForceIndexIndicator.ach = StrategyBasedIndicator.acd(0, ForceIndexIndicator.$);
    return ForceIndexIndicator;
}(StrategyBasedIndicator));
export { ForceIndexIndicator };
