// igniteui-react-charts/esm5/lib/FullStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { FullStochasticOscillatorIndicatorStrategy } from "./FullStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
/**
 * @hidden
 */
var FullStochasticOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FullStochasticOscillatorIndicator, _super);
    function FullStochasticOscillatorIndicator() {
        var _this = _super.call(this) || this;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 10;
        _this.trendLinePeriod = 3;
        return _this;
    }
    FullStochasticOscillatorIndicator.prototype.bj = function () {
        return new FullStochasticOscillatorIndicator();
    };
    FullStochasticOscillatorIndicator.prototype.get_abv = function () {
        return new FullStochasticOscillatorIndicatorStrategy();
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.get_ab8 = function () {
        return FullStochasticOscillatorIndicator.$;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.acl));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.acl, a);
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    FullStochasticOscillatorIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "smoothingPeriod", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.acm));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.acm, a);
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.ab3 = function () {
        return this.smoothingPeriod;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "aci", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.acn));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.acn, a);
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.ab0 = function () {
        return this.aci;
    };
    FullStochasticOscillatorIndicator.prototype.abb = function () {
        return this.aci;
    };
    FullStochasticOscillatorIndicator.$t = markType(FullStochasticOscillatorIndicator, 'FullStochasticOscillatorIndicator', StrategyBasedIndicator.$);
    FullStochasticOscillatorIndicator.acl = StrategyBasedIndicator.acd(14, FullStochasticOscillatorIndicator.$);
    FullStochasticOscillatorIndicator.acm = StrategyBasedIndicator.ace(3, FullStochasticOscillatorIndicator.$, "SmoothingPeriod");
    FullStochasticOscillatorIndicator.acn = StrategyBasedIndicator.ace(3, FullStochasticOscillatorIndicator.$, "TriggerPeriod");
    return FullStochasticOscillatorIndicator;
}(StrategyBasedIndicator));
export { FullStochasticOscillatorIndicator };
