// igniteui-react-charts/esm5/lib/HorizontalRangeCategorySeriesProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, typeCast, markType } from "igniteui-react-core";
import { IHorizontalRangeCategorySeriesProxy_$type } from "./IHorizontalRangeCategorySeriesProxy";
import { HorizontalRangeCategorySeries } from "./HorizontalRangeCategorySeries";
/**
 * @hidden
 */
var HorizontalRangeCategorySeriesProxy = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(HorizontalRangeCategorySeriesProxy, _super);
    function HorizontalRangeCategorySeriesProxy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HorizontalRangeCategorySeriesProxy.prototype.matchesType = function (a) {
        return typeCast(HorizontalRangeCategorySeries.$, a) !== null;
    };
    HorizontalRangeCategorySeriesProxy.prototype.setHighMemberPath = function (a, b) {
        a.aa4 = b;
    };
    HorizontalRangeCategorySeriesProxy.prototype.setLowMemberPath = function (a, b) {
        a.aa7 = b;
    };
    HorizontalRangeCategorySeriesProxy.prototype.setXAxis = function (a, b) {
        a.xAxis = b;
    };
    HorizontalRangeCategorySeriesProxy.prototype.setYAxis = function (a, b) {
        a.yAxis = b;
    };
    HorizontalRangeCategorySeriesProxy.$t = markType(HorizontalRangeCategorySeriesProxy, 'HorizontalRangeCategorySeriesProxy', Base.$, [IHorizontalRangeCategorySeriesProxy_$type]);
    return HorizontalRangeCategorySeriesProxy;
}(Base));
export { HorizontalRangeCategorySeriesProxy };
