// igniteui-react-charts/esm5/lib/HorizontalRangeCategorySeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { RangeCategorySeries } from "./RangeCategorySeries";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { DependencyProperty } from "igniteui-react-core";
import { NumericYAxis } from "./NumericYAxis";
import { Axis } from "./Axis";
import { Point_$type, runOn, typeCast, markType, markDep } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { RangeValueList } from "./RangeValueList";
import { ScalerParams } from "./ScalerParams";
import { CategoryXAxis } from "./CategoryXAxis";
import { DataLegendSeriesValueInfo } from "igniteui-react-core";
import { DataLegendSeriesContext } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
/**
 * @hidden
 */
var HorizontalRangeCategorySeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(HorizontalRangeCategorySeries, _super);
    function HorizontalRangeCategorySeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HorizontalRangeCategorySeries.prototype.fetchXAxis = function () {
        return this.xAxis;
    };
    HorizontalRangeCategorySeries.prototype.fetchYAxis = function () {
        return this.yAxis;
    };
    HorizontalRangeCategorySeries.prototype.o2 = function () {
        _super.prototype.o2.call(this);
        this.xAxis = null;
        this.yAxis = null;
    };
    HorizontalRangeCategorySeries.prototype.aj = function (a) {
        var b = this.an(this.ys(), this.yu.f, this.zv(this.cv), this.zw(this.cv), this.wh(a), true);
        if (b == null) {
            return null;
        }
        var c = b[0];
        var d = b[1];
        var e = new Array(1);
        var f = new Array(2);
        f[0] = { $type: Point_$type, x: c[0], y: c[2] };
        f[1] = { $type: Point_$type, x: d[0], y: d[2] };
        e[0] = f;
        return e;
    };
    HorizontalRangeCategorySeries.prototype.al = function (a) {
        var b = this.an(this.ys(), this.yu.f, this.zv(this.cv), this.zw(this.cv), this.wh(a), true);
        if (b == null) {
            return null;
        }
        var c = b[0];
        var d = b[1];
        var e = new Array(1);
        var f = new Array(2);
        f[0] = { $type: Point_$type, x: c[0], y: c[1] };
        f[1] = { $type: Point_$type, x: d[0], y: d[1] };
        e[0] = f;
        return e;
    };
    HorizontalRangeCategorySeries.prototype.get_fu = function () {
        return true;
    };
    HorizontalRangeCategorySeries.prototype.b2 = function () {
        if (this.b3 == 0) {
            return 1;
        }
        else {
            return _super.prototype.b2.call(this);
        }
    };
    HorizontalRangeCategorySeries.prototype.getOffsetValue = function () {
        return this.RangeFramePreparer.r(this.xAxis, this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
    };
    HorizontalRangeCategorySeries.prototype.getCategoryWidth = function () {
        return this.xAxis.getCategorySize(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
    };
    HorizontalRangeCategorySeries.prototype.j8 = function (a, b) {
        return this.j9(a, b, this.xAxis, runOn(this, this.zt), new RangeValueList(this.highColumn, this.lowColumn));
    };
    HorizontalRangeCategorySeries.prototype.ka = function (a, b) {
        return this.kb(a, b, this.xAxis, runOn(this, this.zt), new RangeValueList(this.highColumn, this.lowColumn));
    };
    HorizontalRangeCategorySeries.prototype.ip = function (a, b, c, d, e) {
        if (c == null) {
            return Number.POSITIVE_INFINITY;
        }
        var f = this.xAxis.lj;
        return this.iq(a, b, this.xAxis, d, e, f, runOn(this, this.zt));
    };
    HorizontalRangeCategorySeries.prototype.i5 = function (a, b, c) {
        if (this.db == null) {
            return NaN;
        }
        var d = this.getEffectiveViewport1(this.cv);
        var e = new ScalerParams(0, this.db.v4, this.cv.bx, this.xAxis.ch, d);
        var f = this.RangeFramePreparer.r(this.xAxis, this.db.v4, this.cv.bx, d);
        return this.i7(new RangeValueList(this.highColumn, this.lowColumn), a, this.xAxis, e, f, runOn(this, this.zt), b, c);
    };
    HorizontalRangeCategorySeries.prototype.i3 = function (a, b, c) {
        if (this.db == null) {
            return NaN;
        }
        var d = this.getEffectiveViewport1(this.cv);
        var e = new ScalerParams(0, this.db.v4, this.cv.bx, this.xAxis.ch, d);
        var f = this.RangeFramePreparer.r(this.xAxis, this.db.v4, this.cv.bx, d);
        return this.i7(this.lowColumn, a, this.xAxis, e, f, runOn(this, this.zt), b, c);
    };
    HorizontalRangeCategorySeries.prototype.i1 = function (a, b, c) {
        if (this.db == null) {
            return NaN;
        }
        var d = this.getEffectiveViewport1(this.cv);
        var e = new ScalerParams(0, this.db.v4, this.cv.bx, this.xAxis.ch, d);
        var f = this.RangeFramePreparer.r(this.xAxis, this.db.v4, this.cv.bx, d);
        return this.i7(this.highColumn, a, this.xAxis, e, f, runOn(this, this.zt), b, c);
    };
    HorizontalRangeCategorySeries.prototype.v8 = function (a, b, c) {
        var _this = this;
        return this.we(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.yAxis, this.xAxis, runOn(this, this.i1), function (d, e) { return _this.kb(d, e, _this.xAxis, runOn(_this, _this.zt), _this.highColumn); }, function (d, e) { return _this.j9(d, e, _this.xAxis, runOn(_this, _this.zt), _this.highColumn); });
    };
    HorizontalRangeCategorySeries.prototype.wa = function (a, b, c) {
        var _this = this;
        return this.we(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.yAxis, this.xAxis, runOn(this, this.i3), function (d, e) { return _this.kb(d, e, _this.xAxis, runOn(_this, _this.zt), _this.lowColumn); }, function (d, e) { return _this.j9(d, e, _this.xAxis, runOn(_this, _this.zt), _this.lowColumn); });
    };
    HorizontalRangeCategorySeries.prototype.wc = function (a, b, c) {
        return this.we(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.yAxis, this.xAxis, null, null, null);
    };
    HorizontalRangeCategorySeries.prototype.zp = function () {
        return this.yAxis != null && this.yAxis.c1();
    };
    HorizontalRangeCategorySeries.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case HorizontalRangeCategorySeries.$$p[0]:
            case HorizontalRangeCategorySeries.$$p[1]:
            case HorizontalRangeCategorySeries.$$p[2]:
            case HorizontalRangeCategorySeries.$$p[3]:
                this.qf();
                break;
            case HorizontalRangeCategorySeries.$$p[4]:
                this.pc(typeCast(Axis.$, c));
                this.rh(typeCast(Axis.$, d));
                this.y3.cw.g(this.id);
                this.rr(false);
                this.ql();
                break;
            case HorizontalRangeCategorySeries.$$p[5]:
                this.pc(typeCast(Axis.$, c));
                this.rh(typeCast(Axis.$, d));
                this.y3.cw.g(this.id);
                this.zp();
                this.rr(false);
                this.ql();
                break;
        }
    };
    HorizontalRangeCategorySeries.prototype.abj = function (a) {
        if (typeCast(NumericYAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    HorizontalRangeCategorySeries.prototype.abi = function (a) {
        if (typeCast(CategoryXAxis.$, a) !== null || a.ce) {
            return true;
        }
        return false;
    };
    HorizontalRangeCategorySeries.prototype.getDataLegendSeriesContextAt = function (a, b) {
        var c = { $type: Point_$type, x: a, y: b };
        var d = this.i3(c, false, false);
        var e = this.i1(c, false, false);
        var f = new DataLegendSeriesValueInfo();
        f.a = 1;
        f.l = e > d ? d : e;
        f.q = "Low";
        f.r = this.aa7 != null ? this.aa7 : "Low";
        f.s = this.abo != null ? this.abo : "L:";
        f.t = this.abq != null ? this.abq : "";
        var g = new DataLegendSeriesValueInfo();
        g.a = 2;
        g.l = e > d ? e : d;
        g.q = "High";
        g.r = this.aa4 != null ? this.aa4 : "High";
        g.s = this.abk != null ? this.abk : "H:";
        g.t = this.abm != null ? this.abm : "";
        var h = new DataLegendSeriesValueInfo();
        h.a = 7;
        h.l = Math.abs(g.l - f.l);
        h.q = "Range";
        h.r = "Range";
        h.s = "R:";
        h.k = true;
        var i = new DataLegendSeriesValueInfo();
        i.a = 6;
        i.l = f.l + (h.l / 2);
        i.q = "Average";
        i.r = "Average";
        i.s = "A:";
        i.k = true;
        var j = new DataLegendSeriesContext();
        j.e = ((function () {
            var $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
            $ret.add(g);
            $ret.add(f);
            $ret.add(h);
            $ret.add(i);
            return $ret;
        })());
        j.h = this.mc();
        return j;
    };
    Object.defineProperty(HorizontalRangeCategorySeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zq;
        },
        enumerable: false,
        configurable: true
    });
    HorizontalRangeCategorySeries.$t = markType(HorizontalRangeCategorySeries, 'HorizontalRangeCategorySeries', RangeCategorySeries.$);
    HorizontalRangeCategorySeries.$$p = markDep(DependencyProperty, PropertyMetadata, HorizontalRangeCategorySeries, 'raisePropertyChanged', ['HighMemberAsLegendLabel:abk:abu', [2, null], 'HighMemberAsLegendUnit:abm:abv', [2, null], 'LowMemberAsLegendLabel:abo:abw', [2, null], 'LowMemberAsLegendUnit:abq:abx', [2, null], 'XAxis::aby', [CategoryAxisBase.$, null], 'YAxis::abz', [NumericYAxis.$, null]]);
    return HorizontalRangeCategorySeries;
}(RangeCategorySeries));
export { HorizontalRangeCategorySeries };
