// igniteui-react-charts/esm5/lib/igr-anchored-radial-series.js
import { __extends } from "tslib";
import { TrendLineType_$type } from "igniteui-react-core";
import { IgrRadialBase } from "./igr-radial-base";
import { ensureBool, ensureEnum, brushToString, stringToBrush, toDoubleCollection, fromDoubleCollection, toPoint, fromPoint } from "igniteui-react-core";
/**
 * Represents the base class for all IgxDataChartComponent anchored radial category series.
*/
var IgrAnchoredRadialSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgrAnchoredRadialSeries, _super);
    function IgrAnchoredRadialSeries(props) {
        return _super.call(this, props) || this;
    }
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "useCategoryNormalizedValues", {
        /**
         * Gets or sets whether to normalize the values against the category values if using a proportional category axis.
        */
        get: function () {
            return this.i.ach;
        },
        set: function (v) {
            this.i.ach = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "valueMemberPath", {
        /**
         * Gets or sets the item path that provides the values for the current series.
        */
        get: function () {
            return this.i.acx;
        },
        set: function (v) {
            this.i.acx = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "highlightedValueMemberPath", {
        /**
         * Gets or sets the value mapping property for the current series object.
        */
        get: function () {
            return this.i.acp;
        },
        set: function (v) {
            this.i.acp = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "valueMemberAsLegendLabel", {
        /**
         * Gets or sets the label displayed before series value in the Data Legend.
        */
        get: function () {
            return this.i.act;
        },
        set: function (v) {
            this.i.act = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "valueMemberAsLegendUnit", {
        /**
         * Gets or sets the unit displayed after series value in the Data Legend.
        */
        get: function () {
            return this.i.acv;
        },
        set: function (v) {
            this.i.acv = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "trendLineType", {
        /**
         * Gets or sets the trend type for the current series object.
        */
        get: function () {
            return this.i.ab6;
        },
        set: function (v) {
            this.i.ab6 = ensureEnum(TrendLineType_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "trendLineBrush", {
        /**
         * Gets or sets the brush that specifies how to the current series
         * object's Trend line is drawn.
        */
        get: function () {
            return brushToString(this.i.adl);
        },
        set: function (v) {
            this.i.adl = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "actualTrendLineBrush", {
        /**
         * Gets the effective TrendLineBrush for this series.
        */
        get: function () {
            return brushToString(this.i.adk);
        },
        set: function (v) {
            this.i.adk = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "trendLineThickness", {
        /**
         * Gets or sets the thickness of the current series object's trend line.
        */
        get: function () {
            return this.i.acm;
        },
        set: function (v) {
            this.i.acm = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "trendLineDashArray", {
        /**
         * Gets or sets a collection of double values that indicate the pattern of dashes and gaps that
         * is used to draw the trend line for the current series object.
        */
        get: function () {
            return fromDoubleCollection(this.i.adm);
        },
        set: function (v) {
            this.i.adm = toDoubleCollection(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "trendLinePeriod", {
        /**
         * Gets or sets the trend line period for the current series.
         * The typical, and initial, value for bollinger band periods is 20.
        */
        get: function () {
            return this.i.acn;
        },
        set: function (v) {
            this.i.acn = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "trendLineZIndex", {
        /**
         * Sets or Gets the Z index of the trendline.
        */
        get: function () {
            return this.i.aco;
        },
        set: function (v) {
            this.i.aco = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "hasValueAxis", {
        get: function () {
            return this.i.ek;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrAnchoredRadialSeries.prototype, "isValueAxisInverted", {
        get: function () {
            return this.i.f5;
        },
        enumerable: false,
        configurable: true
    });
    IgrAnchoredRadialSeries.prototype.getItemValue = function (item, memberPathName) {
        var iv = this.i.kq(item, memberPathName);
        return (iv);
    };
    /**
     * Gets the value of a requested member path from the series.
    
    * @param memberPathName  * The property name of a valid member path for the series
    */
    IgrAnchoredRadialSeries.prototype.getMemberPathValue = function (memberPathName) {
        var iv = this.i.md(memberPathName);
        return (iv);
    };
    IgrAnchoredRadialSeries.prototype.getPreviousOrExactIndex = function (world, skipUnknowns) {
        var iv = this.i.ka(toPoint(world), skipUnknowns);
        return (iv);
    };
    IgrAnchoredRadialSeries.prototype.getNextOrExactIndex = function (world, skipUnknowns) {
        var iv = this.i.j8(toPoint(world), skipUnknowns);
        return (iv);
    };
    IgrAnchoredRadialSeries.prototype.getSeriesValue = function (world, useInterpolation, skipUnknowns) {
        var iv = this.i.i5(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    };
    IgrAnchoredRadialSeries.prototype.getSeriesValuePosition = function (world, useInterpolation, skipUnknowns) {
        var iv = this.i.wc(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    };
    /**
     * Returns the offset value for this series if grouped on a category axis.
    
    */
    IgrAnchoredRadialSeries.prototype.getOffsetValue = function () {
        var iv = this.i.getOffsetValue();
        return (iv);
    };
    /**
     * Returns the width of the category grouping this series is in.
    
    */
    IgrAnchoredRadialSeries.prototype.getCategoryWidth = function () {
        var iv = this.i.getCategoryWidth();
        return (iv);
    };
    /**
     * Scrolls the specified item into the view.
    
    * @param item  * The item to scroll into view.
    */
    IgrAnchoredRadialSeries.prototype.scrollIntoView = function (item) {
        var iv = this.i.gd(item);
        return (iv);
    };
    return IgrAnchoredRadialSeries;
}(IgrRadialBase));
export { IgrAnchoredRadialSeries };
