// igniteui-react-charts/esm5/lib/igr-chart-summary-description-collection.js
import { __generator } from "tslib";
import { ensureBool } from "igniteui-react-core";
import { IgrChartSummaryDescription } from "./igr-chart-summary-description";
import { ChartSummaryDescriptionCollection as ChartSummaryDescriptionCollection_internal } from "./ChartSummaryDescriptionCollection";
var IgrChartSummaryDescriptionCollection = /** @class */ /*@__PURE__*/ (function () {
    function IgrChartSummaryDescriptionCollection() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    IgrChartSummaryDescriptionCollection.prototype.createImplementation = function () {
        return new ChartSummaryDescriptionCollection_internal();
    };
    Object.defineProperty(IgrChartSummaryDescriptionCollection.prototype, "i", {
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrChartSummaryDescriptionCollection.prototype.onImplementationCreated = function () {
    };
    IgrChartSummaryDescriptionCollection.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    };
    IgrChartSummaryDescriptionCollection.prototype.item = function (index, value) {
        var int = null;
        if (value !== undefined) {
            int = this.i.set(index, value == null ? null : value.i);
        }
        else {
            int = this.i.get(index);
        }
        var ext = null;
        if (int != null) {
            ext = int.externalObject;
            if (!ext) {
                ext = new IgrChartSummaryDescription();
                ext._implementation = int;
                int.externalObject = ext;
            }
        }
        return ext;
    };
    Object.defineProperty(IgrChartSummaryDescriptionCollection.prototype, "count", {
        get: function () {
            return this.i.size();
        },
        enumerable: false,
        configurable: true
    });
    IgrChartSummaryDescriptionCollection.prototype.toArray = function () {
        var arr = [];
        for (var i = 0; i < this.count; i++) {
            arr[i] = this.item(i);
        }
        return arr;
    };
    IgrChartSummaryDescriptionCollection.prototype[Symbol.iterator] = function () {
        var i, item;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    i = 0;
                    _a.label = 1;
                case 1:
                    if (!(i < this.count))
                        return [3 /*break*/, 4];
                    item = this.item(i);
                    if (item.externalObject) {
                        item = item.externalObject;
                    }
                    return [4 /*yield*/, item];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3:
                    i++;
                    return [3 /*break*/, 1];
                case 4: return [2 /*return*/];
            }
        });
    };
    Object.defineProperty(IgrChartSummaryDescriptionCollection.prototype, "shouldDetachOnTargetChange", {
        /**
     * Gets or sets whether this collection should detach the sync when the target collection changes.
    */
        get: function () {
            return this.i.i;
        },
        set: function (v) {
            this.i.i = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    IgrChartSummaryDescriptionCollection.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    IgrChartSummaryDescriptionCollection.prototype.add = function (item) {
        var iv = this.i.g((item == null ? null : item.i));
        return (iv);
    };
    IgrChartSummaryDescriptionCollection.prototype.insert = function (index, item) {
        this.i.p(index, (item == null ? null : item.i));
    };
    IgrChartSummaryDescriptionCollection.prototype.clear = function () {
        this.i.o();
    };
    IgrChartSummaryDescriptionCollection.prototype.indexOf = function (item) {
        var iv = this.i.k((item == null ? null : item.i));
        return (iv);
    };
    IgrChartSummaryDescriptionCollection.prototype.remove = function (item) {
        var iv = this.i.h((item == null ? null : item.i));
        return (iv);
    };
    IgrChartSummaryDescriptionCollection.prototype.removeAt = function (index) {
        var iv = this.i.b(index);
        var ret = null;
        if (iv && iv.externalObject) {
            ret = iv.externalObject;
        }
        else {
            if (iv) {
                var e = new IgrChartSummaryDescription();
                e._implementation = iv;
                iv.externalObject = e;
                ret = e;
            }
        }
        return ret;
    };
    return IgrChartSummaryDescriptionCollection;
}());
export { IgrChartSummaryDescriptionCollection };
