// igniteui-react-charts/esm5/lib/igr-series-viewer.js
import { __extends, __values } from "tslib";
import * as React from 'react';
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { MarkerAutomaticBehavior_$type } from "./MarkerAutomaticBehavior";
import { SeriesHighlightingMode_$type } from "./SeriesHighlightingMode";
import { SeriesSelectionMode_$type } from "./SeriesSelectionMode";
import { SeriesSelectionBehavior_$type } from "./SeriesSelectionBehavior";
import { SeriesViewerScrollbarMode_$type } from "./SeriesViewerScrollbarMode";
import { SeriesViewerHorizontalScrollbarPosition_$type } from "./SeriesViewerHorizontalScrollbarPosition";
import { SeriesViewerVerticalScrollbarPosition_$type } from "./SeriesViewerVerticalScrollbarPosition";
import { SeriesHighlightingBehavior_$type } from "./SeriesHighlightingBehavior";
import { LegendHighlightingMode_$type } from "./LegendHighlightingMode";
import { LegendItemBadgeShape_$type } from "igniteui-react-core";
import { LegendItemBadgeMode_$type } from "igniteui-react-core";
import { ChartHitTestMode_$type } from "./ChartHitTestMode";
import { WindowResponse_$type } from "./WindowResponse";
import { Visibility_$type } from "igniteui-react-core";
import { ZoomCoercionMode_$type } from "./ZoomCoercionMode";
import { InteractionState_$type } from "igniteui-react-core";
import { ModifierKeys_$type } from "igniteui-react-core";
import { SeriesHighlightedValuesDisplayMode_$type } from "igniteui-react-core";
import { IgrChartSelectedItemCollection } from "./igr-chart-selected-item-collection";
import { HorizontalAlignment_$type } from "igniteui-react-core";
import { IgrPlotAreaMouseButtonEventArgs } from "./igr-plot-area-mouse-button-event-args";
import { IgrPlotAreaMouseEventArgs } from "./igr-plot-area-mouse-event-args";
import { IgrChartCursorEventArgs } from "./igr-chart-cursor-event-args";
import { IgrDataChartMouseButtonEventArgs } from "./igr-data-chart-mouse-button-event-args";
import { IgrChartMouseEventArgs } from "./igr-chart-mouse-event-args";
import { IgrSeriesViewerManipulationEventArgs } from "./igr-series-viewer-manipulation-event-args";
import { IgrRectChangedEventArgs } from "igniteui-react-core";
import { IgrSeriesViewerSelectedSeriesItemsChangedEventArgs } from "./igr-series-viewer-selected-series-items-changed-event-args";
import { IgrRefreshCompletedEventArgs } from "./igr-refresh-completed-event-args";
import { IgrImageCapturedEventArgs } from "igniteui-react-core";
import { TypeRegistrar } from "igniteui-react-core";
import { NamePatcher, getModifiedProps, isValidProp, toBrushCollection, fromBrushCollection, ensureEnum, fromPoint, toPoint, ensureBool, brushToString, stringToBrush, fromRect, toRect, toSpinal, initializePropertiesFromCss } from "igniteui-react-core";
import { ChartSelectedItemCollection as ChartSelectedItemCollection_internal } from "./ChartSelectedItemCollection";
import { ChartSelection } from "./ChartSelection";
import { SyncableObservableCollection$1 } from "igniteui-react-core";
/**
 * Represents a control for viewing series objects.
*/
var IgrSeriesViewer = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgrSeriesViewer, _super);
    function IgrSeriesViewer(props) {
        var _this = _super.call(this, props) || this;
        _this.mounted = false;
        _this._selectedSeriesItems = null;
        _this._focusedSeriesItems = null;
        _this.__p = null;
        _this._hasUserValues = new Set();
        _this._stylingContainer = null;
        _this._stylingParent = null;
        _this._inStyling = false;
        _this._plotAreaMouseLeftButtonDown = null;
        _this._plotAreaMouseLeftButtonDown_wrapped = null;
        _this._plotAreaMouseLeftButtonUp = null;
        _this._plotAreaMouseLeftButtonUp_wrapped = null;
        _this._plotAreaClicked = null;
        _this._plotAreaClicked_wrapped = null;
        _this._plotAreaMouseEnter = null;
        _this._plotAreaMouseEnter_wrapped = null;
        _this._plotAreaMouseLeave = null;
        _this._plotAreaMouseLeave_wrapped = null;
        _this._plotAreaMouseOver = null;
        _this._plotAreaMouseOver_wrapped = null;
        _this._seriesCursorMouseMove = null;
        _this._seriesCursorMouseMove_wrapped = null;
        _this._seriesMouseLeftButtonDown = null;
        _this._seriesMouseLeftButtonDown_wrapped = null;
        _this._seriesMouseLeftButtonUp = null;
        _this._seriesMouseLeftButtonUp_wrapped = null;
        _this._seriesClick = null;
        _this._seriesClick_wrapped = null;
        _this._seriesMouseMove = null;
        _this._seriesMouseMove_wrapped = null;
        _this._seriesMouseEnter = null;
        _this._seriesMouseEnter_wrapped = null;
        _this._seriesMouseLeave = null;
        _this._seriesMouseLeave_wrapped = null;
        _this._viewerManipulationStarting = null;
        _this._viewerManipulationStarting_wrapped = null;
        _this._viewerManipulationEnding = null;
        _this._viewerManipulationEnding_wrapped = null;
        _this._windowRectChanged = null;
        _this._windowRectChanged_wrapped = null;
        _this._sizeChanged = null;
        _this._sizeChanged_wrapped = null;
        _this._actualWindowRectChanged = null;
        _this._actualWindowRectChanged_wrapped = null;
        _this._gridAreaRectChanged = null;
        _this._gridAreaRectChanged_wrapped = null;
        _this._selectedSeriesItemsChanged = null;
        _this._selectedSeriesItemsChanged_wrapped = null;
        _this._focusedSeriesItemsChanged = null;
        _this._focusedSeriesItemsChanged_wrapped = null;
        _this._refreshCompleted = null;
        _this._refreshCompleted_wrapped = null;
        _this._imageCaptured = null;
        _this._imageCaptured_wrapped = null;
        if (_this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(_this));
        }
        _this._implementation = _this.createImplementation();
        _this._implementation.externalObject = _this;
        _this.onImplementationCreated();
        if (_this._initializeAdapters) {
            _this._initializeAdapters();
        }
        return _this;
    }
    IgrSeriesViewer.prototype.createImplementation = function () {
        return null;
    };
    Object.defineProperty(IgrSeriesViewer.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        } /**
                                     * @hidden
                                     */,
        enumerable: false,
        configurable: true
    });
    IgrSeriesViewer._createFromInternal = function (internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        var name = internal.$type.name;
        var externalName = "Igr" + name;
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    };
    IgrSeriesViewer.prototype.onImplementationCreated = function () {
    };
    IgrSeriesViewer.prototype.componentDidMount = function () {
        var e_1, _a;
        this.mounted = true;
        try {
            for (var _b = __values(Object.keys(this.props)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var p = _c.value;
                if (isValidProp(this, p)) {
                    this[p] = this.props[p];
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
    };
    IgrSeriesViewer.prototype.shouldComponentUpdate = function (nextProps, nextState) {
        var e_2, _a;
        var mod = getModifiedProps(this.props, nextProps);
        try {
            for (var _b = __values(Object.keys(mod)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var p = _c.value;
                if (isValidProp(this, p)) {
                    this[p] = mod[p];
                }
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
        return true;
    };
    IgrSeriesViewer.prototype.render = function () {
        return null;
    };
    Object.defineProperty(IgrSeriesViewer.prototype, "brushes", {
        /**
         * Gets or sets the palette of brushes used for fill of plotted series.
         * These brushes are automatically assigned to series based on the index of series.
        */
        get: function () {
            return fromBrushCollection(this.i.ag);
        },
        set: function (v) {
            this.i.ag = toBrushCollection(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "outlines", {
        /**
         * Gets or sets the palette of brushes used for outlines of plotted series.
         * These outlines are automatically assigned to series based on the index of series.
        */
        get: function () {
            return fromBrushCollection(this.i.aj);
        },
        set: function (v) {
            this.i.aj = toBrushCollection(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "markerBrushes", {
        /**
         * Gets or sets the palette of brushes used for fill of marker series.
         * These brushes are automatically assigned to markers based on the index of series.
        */
        get: function () {
            return fromBrushCollection(this.i.ah);
        },
        set: function (v) {
            this.i.ah = toBrushCollection(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "markerOutlines", {
        /**
         * Gets or sets the palette of brushes used for outlines of plotted markers.
         * These outlines are automatically assigned to markers based on the index of series.
        */
        get: function () {
            return fromBrushCollection(this.i.ai);
        },
        set: function (v) {
            this.i.ai = toBrushCollection(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "markerAutomaticBehavior", {
        /**
         * Gets or sets the marker mode used for assigning a marker type to series when the marker type is automatic.
        */
        get: function () {
            return this.i.bg;
        },
        set: function (v) {
            this.i.bg = ensureEnum(MarkerAutomaticBehavior_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "crosshairPoint", {
        /**
         * Gets or sets the cross hair point (in world coordinates)
         * Either or both of the crosshair point's X and Y may be set to double.NaN, in which
         * case the relevant crosshair line is hidden.
        *
        * The `CrosshairPoint` property is used to gets/sets the cross hair point (in world coordinates).
        *
        * ```ts
        * this.chart.crosshairPoint = {x:.8,y:.2};
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	crosshairVisibility="visible"
        *     crosshairPoint={x:.8,y:.2} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrAreaSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value"
        *     brush="Gray"
        *     outline="Black" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return fromPoint(this.i.vv);
        },
        set: function (v) {
            this.i.vv = toPoint(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "animateSeriesWhenAxisRangeChanges", {
        /**
         * Gets or sets whether the series animations should be allowed when a range change has been detected on an axis.
        *
        * `AnimateSeriesWhenAxisRangeChanges` is a Boolean property to override the default behavior in which series do not animate if an axis range changes.
        *
        * ```ts
        * this.chart.animateSeriesWhenAxisRangeChanges = true;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	animateSeriesWhenAxisRangeChanges={true}>
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrColumnSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.ek;
        },
        set: function (v) {
            this.i.ek = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "seriesValueLayerUseLegend", {
        /**
         * Gets or sets whether the series animations should be allowed when a range change has been detected on an axis.
        */
        get: function () {
            return this.i.f2;
        },
        set: function (v) {
            this.i.f2 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "isWindowSyncedToVisibleRange", {
        /**
         * Gets or sets whether to sync the vertical aspect of the window with the contained series content. This is only supported for a subset of the available series.
        */
        get: function () {
            return this.i.fj;
        },
        set: function (v) {
            this.i.fj = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "shouldSimulateHoverMoveCrosshairPoint", {
        /**
         * Gets or sets whether calling SimulateHover should shift the crosshair point.
        */
        get: function () {
            return this.i.f7;
        },
        set: function (v) {
            this.i.f7 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "highlightingMode", {
        /**
         * Gets or sets the highlighting mode to use for the series in the component, when supported. This takes precedence over the series level IsHighlightingEnabled.
        */
        get: function () {
            return this.i.by;
        },
        set: function (v) {
            this.i.by = ensureEnum(SeriesHighlightingMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "selectionMode", {
        /**
         * Gets or sets the selection mode to use for the series in the component, when supported.
        */
        get: function () {
            return this.i.cb;
        },
        set: function (v) {
            this.i.cb = ensureEnum(SeriesSelectionMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "selectionBehavior", {
        /**
         * Gets or sets the selection behavior to use for the series in the component, when supported.
        */
        get: function () {
            return this.i.b6;
        },
        set: function (v) {
            this.i.b6 = ensureEnum(SeriesSelectionBehavior_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "focusMode", {
        /**
         * Gets or sets the focus mode to use for the series in the component, when supported.
        */
        get: function () {
            return this.i.b9;
        },
        set: function (v) {
            this.i.b9 = ensureEnum(SeriesSelectionMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarMode", {
        /**
         * Gets or sets the horizontal scrollbar mode to use for the series viewer.
        */
        get: function () {
            return this.i.ch;
        },
        set: function (v) {
            this.i.ch = ensureEnum(SeriesViewerScrollbarMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarMode", {
        /**
         * Gets or sets the vertical scrollbar mode to use for the series viewer.
        */
        get: function () {
            return this.i.ci;
        },
        set: function (v) {
            this.i.ci = ensureEnum(SeriesViewerScrollbarMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarPosition", {
        /**
         * Gets or sets the position to use for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.ce;
        },
        set: function (v) {
            this.i.ce = ensureEnum(SeriesViewerHorizontalScrollbarPosition_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarPosition", {
        /**
         * Gets or sets the position to use for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.cl;
        },
        set: function (v) {
            this.i.cl = ensureEnum(SeriesViewerVerticalScrollbarPosition_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarFill", {
        /**
         * Gets or sets the fill to use for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return brushToString(this.i.vi);
        },
        set: function (v) {
            this.i.vi = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarOutline", {
        /**
         * Gets or sets the outline to use for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return brushToString(this.i.vj);
        },
        set: function (v) {
            this.i.vj = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarStrokeThickness", {
        /**
         * Gets or sets the stroke thickness to use for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.hy;
        },
        set: function (v) {
            this.i.hy = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarMaxOpacity", {
        /**
         * Gets or sets the max opacity to use for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.hx;
        },
        set: function (v) {
            this.i.hx = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarCornerRadius", {
        /**
         * Gets or sets the corner radius to use for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.hu;
        },
        set: function (v) {
            this.i.hu = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarHeight", {
        /**
         * Gets or sets the height to use for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.hv;
        },
        set: function (v) {
            this.i.hv = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarInset", {
        /**
         * Gets or sets the inset distance to use for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.hw;
        },
        set: function (v) {
            this.i.hw = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarTrackStartInset", {
        /**
         * Gets or sets the track start inset to use for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.h0;
        },
        set: function (v) {
            this.i.h0 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarTrackEndInset", {
        /**
         * Gets or sets the track end inset to use for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.hz;
        },
        set: function (v) {
            this.i.hz = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalViewScrollbarShouldAddAutoTrackInsets", {
        /**
         * Gets or sets whether to use automatic track insets for the horizontal scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.ey;
        },
        set: function (v) {
            this.i.ey = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarFill", {
        /**
         * Gets or sets the fill to use for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return brushToString(this.i.vs);
        },
        set: function (v) {
            this.i.vs = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarOutline", {
        /**
         * Gets or sets the outline to use for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return brushToString(this.i.vt);
        },
        set: function (v) {
            this.i.vt = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarStrokeThickness", {
        /**
         * Gets or sets the stroke thickness to use for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.il;
        },
        set: function (v) {
            this.i.il = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarMaxOpacity", {
        /**
         * Gets or sets the max opacity to use for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.ik;
        },
        set: function (v) {
            this.i.ik = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarCornerRadius", {
        /**
         * Gets or sets the corner radius to use for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.ii;
        },
        set: function (v) {
            this.i.ii = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarWidth", {
        /**
         * Gets or sets the width to use for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.ip;
        },
        set: function (v) {
            this.i.ip = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarInset", {
        /**
         * Gets or sets the inset distance to use for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.ij;
        },
        set: function (v) {
            this.i.ij = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarTrackStartInset", {
        /**
         * Gets or sets the track start inset to use for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.io;
        },
        set: function (v) {
            this.i.io = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarTrackEndInset", {
        /**
         * Gets or sets the track end inset to use for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.im;
        },
        set: function (v) {
            this.i.im = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalViewScrollbarShouldAddAutoTrackInsets", {
        /**
         * Gets or sets whether to use automatic track insets for the vertical scrollbar in the component, when enabled.
        */
        get: function () {
            return this.i.gj;
        },
        set: function (v) {
            this.i.gj = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "highlightingBehavior", {
        /**
         * Gets or sets the highlighting Behavior to use for the series in the component, when supported. This takes precedence over the series level IsHighlightingEnabled.
        */
        get: function () {
            return this.i.bw;
        },
        set: function (v) {
            this.i.bw = ensureEnum(SeriesHighlightingBehavior_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "legendHighlightingMode", {
        /**
         * Gets or sets the highlighting mode to use for the legend linked to the component, when supported.
        */
        get: function () {
            return this.i.bd;
        },
        set: function (v) {
            this.i.bd = ensureEnum(LegendHighlightingMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "legendItemBadgeShape", {
        /**
         * Gets or sets type of legend badges representing all series displayed in a legend linked to this component
         * This property will be ignored when the LegendItemTemplate or LegendItemBadgeTemplate property is set on a series
        */
        get: function () {
            return this.i.bf;
        },
        set: function (v) {
            this.i.bf = ensureEnum(LegendItemBadgeShape_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "legendItemBadgeMode", {
        /**
         * Gets or sets the mode of legend badges representing all series in this chart.
         * This property will be ignored when the LegendItemTemplate or LegendItemBadgeTemplate property is set on a series
        */
        get: function () {
            return this.i.be;
        },
        set: function (v) {
            this.i.be = ensureEnum(LegendItemBadgeMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "viewportRect", {
        /**
         * Gets the viewport rectangle associated with the SeriesViewer, the physical dimensions of the plot area.
        */
        get: function () {
            return fromRect(this.i.wk);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "effectiveViewport", {
        /**
         * Gets the EffectiveViewport rectangle, representing the effective viewport area after margins have been subtracted.
        *
        *  ``ts
        * let viewport: IgxRect = this.chart.effectiveViewport;
        *
        * ```
        *
        * ```
        */
        get: function () {
            return fromRect(this.i.v8);
        },
        set: function (v) {
            this.i.v8 = toRect(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "contentViewport", {
        /**
         * Represents the subset area within the viewport into which content should be mapped.
         * This is the projection of the current EffectiveViewport into the viewport at current zoom level.
        */
        get: function () {
            return fromRect(this.i.v7);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "windowRect", {
        /**
         * A rectangle representing the portion of the SeriesViewer currently in view.
         * A rectangle at X=0, Y=0 with a Height and Width of 1 implies the entire plotting area is in view.  A Height and Width of .5 would imply that the view is halfway zoomed in.
        *
        * You can set the `WindowRect` to zoom in on an area of interest in the chart.
        *
        * ```ts
        * this.chart.windowRect = {left:0, top:1, width: .5, height: .5};
        * ```
        */
        get: function () {
            return fromRect(this.i.wl);
        },
        set: function (v) {
            this.i.wl = toRect(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "useTiledZooming", {
        /**
         * Gets or sets whether the series viewer should use cached tiles during zooms rather than the default live content.
        */
        get: function () {
            return this.i.gi;
        },
        set: function (v) {
            this.i.gi = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "preferHigherResolutionTiles", {
        /**
         * Gets or sets whether the series viewer should prefer selecting higher resolution tiles over lower resolution tiles when performing tile zooming. Setting this to true will lower performance but increase quality.
        */
        get: function () {
            return this.i.fz;
        },
        set: function (v) {
            this.i.fz = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "zoomTileCacheSize", {
        /**
         * Gets or sets the maximum number of zoom tiles that the series viewer should cache while in tiled zooming mode.
        */
        get: function () {
            return this.i.i8;
        },
        set: function (v) {
            this.i.i8 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "highlightingTransitionDuration", {
        /**
         * Gets or sets the duration the highlighting/de-highlighting effects.
        *
        * The `HighlightingTransitionDuration` property is used to set the time for highlighting transition in milliseconds .
        *
        * ```ts
        * chart.HighlightingTransitionDuration=500;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	HighlightingTransitionDuration={500} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrAreaSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value"
        *     brush="Gray"
        *     outline="Black" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.i5;
        },
        set: function (v) {
            this.i.i5 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "selectionTransitionDuration", {
        /**
         * Gets or sets the duration the selection/de-Selection effects.
        */
        get: function () {
            return this.i.i7;
        },
        set: function (v) {
            this.i.i7 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "focusTransitionDuration", {
        /**
         * Gets or sets the duration the Focus/de-Focus effects.
        */
        get: function () {
            return this.i.i3;
        },
        set: function (v) {
            this.i.i3 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "scrollbarsAnimationDuration", {
        /**
         * Gets or sets the duration of the scrollbar effects.
        */
        get: function () {
            return this.i.i6;
        },
        set: function (v) {
            this.i.i6 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "isPagePanningAllowed", {
        /**
         * Gets or sets the whether the series viewer can allow the page to pan if a control pan is not possible in the requested direction.
        */
        get: function () {
            return this.i.fe;
        },
        set: function (v) {
            this.i.fe = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "contentHitTestMode", {
        /**
         * Determines the method of hit testing used when mousing over elements in the chart.
        *
        * ```ts
        * this.chart.contentHitTestMode = ChartHitTestMode.Computational;
        * ```
        */
        get: function () {
            return this.i.av;
        },
        set: function (v) {
            this.i.av = ensureEnum(ChartHitTestMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "actualContentHitTestMode", {
        /**
         * Resolved method of hit testing used when the pointer is over elements in the chart.
        *
        * ```ts
        * let actualContentHitTestMode: ChartHitTestMode  =this.chart.actualContentHitTestMode;
        * ```
        */
        get: function () {
            return this.i.au;
        },
        set: function (v) {
            this.i.au = ensureEnum(ChartHitTestMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "isSurfaceInteractionDisabled", {
        /**
         * Gets or sets whether all surface interactions with the plot area should be disabled.
        *
        * The `IsSurfaceInteractionDisabled` property is used to enable/disable interactions with the plot surface.
        *
        * ```ts
        * chart.isSurfaceInteractionDisabled=true;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	 isSurfaceInteractionDisabled={true} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrAreaSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value"
        *     brush="Gray"
        *     outline="Black" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.ff;
        },
        set: function (v) {
            this.i.ff = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "isMap", {
        /**
         * Indicates if this SeriesViewer is a map.
        */
        get: function () {
            return this.i.fd;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "windowResponse", {
        /**
         * The response to user panning and zooming: whether to update the view immediately while the user action is happening, or to defer the update to after the user action is complete.  The user action will be an action such as a mouse drag which causes panning and/or zooming to occur.
        *
        * The `WindowResponse` property is used to set the response to user panning and zooming: whether to update the view immediately while the user action is happening, or to defer the update to after the user action is complete.
        * The user action will be an action such as a mouse drag which causes panning and/or zooming to occur.
        *
        * ```ts
        * this.chart.windowResponse="deferred";
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	windowResponse="deferred">
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.cp;
        },
        set: function (v) {
            this.i.cp = ensureEnum(WindowResponse_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "windowRectMinWidth", {
        /**
         * Sets or gets the minimum width that the window rect is allowed to reach before being clamped.
         * Decrease this value if you want to allow for further zooming into the viewer.
         * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
         * This property accepts values between 0.0 (max zooming) and 1.0 (no zooming)
        *
        * The `WindowRectMinWidth` property is used to set or get the minimum width that the window rect is allowed to reach before being clamped.
        *
        * ```ts
        * this.chart.WindowRectMinWidth=".2";
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	WindowRectMinWidth=".2">
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.it;
        },
        set: function (v) {
            this.i.it = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "windowRectMinHeight", {
        /**
         * Sets or gets the minimum height that the window rect is allowed to reach before being clamped.
         * Decrease this value if you want to allow for further zooming into the viewer.
         * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
         * This property accepts values between 0.0 (max zooming) and 1.0 (no zooming)
        */
        get: function () {
            return this.i.is;
        },
        set: function (v) {
            this.i.is = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "actualWindowRectMinWidth", {
        /**
         * Gets the actual minimum width that the window rect is allowed to reach before being clamped.
        */
        get: function () {
            return this.i.hm;
        },
        set: function (v) {
            this.i.hm = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "actualWindowRectMinHeight", {
        /**
         * Gets the actual minimum height that the window rect is allowed to reach before being clamped.
        */
        get: function () {
            return this.i.hl;
        },
        set: function (v) {
            this.i.hl = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "windowSizeMinWidth", {
        /**
         * Sets or gets minimum pixel width that the window is allowed to reach before being clamped.
         * Decrease this value if you want to allow for further zooming into the viewer.
         * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
         * This property is overridden by the WindowRectMinWidth property
        */
        get: function () {
            return this.i.iv;
        },
        set: function (v) {
            this.i.iv = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "windowSizeMinHeight", {
        /**
         * Sets or gets minimum pixel height that the window is allowed to reach before being clamped.
         * Decrease this value if you want to allow for further zooming into the viewer.
         * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
         * This property is overridden by the WindowRectMinHeight property
        */
        get: function () {
            return this.i.iu;
        },
        set: function (v) {
            this.i.iu = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "crosshairVisibility", {
        /**
         * Gets or sets the current SeriesViewer's crosshair visibility override.
        *
        * The `CrosshairVisibility` property is used to gets or sets the current Chart's crosshair visibility override. Note: setting this property does not affect the mobile browser version of the chart.
        *
        * ```ts
        * this.chart.crosshairVisibility = Visibility.Visible;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	crosshairVisibility="visible"
        *     crosshairPoint={x:.8,y:.2} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrAreaSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value"
        *     brush="Gray"
        *     outline="Black" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.wp;
        },
        set: function (v) {
            this.i.wp = ensureEnum(Visibility_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "horizontalCrosshairBrush", {
        /**
         * Gets or sets the current SeriesViewer's crosshair horizontal brush override.
        */
        get: function () {
            return brushToString(this.i.vh);
        },
        set: function (v) {
            this.i.vh = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "verticalCrosshairBrush", {
        /**
         * Gets or sets the current SeriesViewer's crosshair vertical brush override.
        */
        get: function () {
            return brushToString(this.i.vr);
        },
        set: function (v) {
            this.i.vr = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "zoomCoercionMode", {
        /**
         * Gets or sets zoom is constrained to within the axes. Setting this to false is a preview feature at the present time.
        */
        get: function () {
            return this.i.ct;
        },
        set: function (v) {
            this.i.ct = ensureEnum(ZoomCoercionMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "plotAreaBackground", {
        /**
         * Gets or sets the brush used as the background for the current SeriesViewer object's plot area.
        *
        * `PlotAreaBackground` property set the background brush to the current Chart object's plot area.
        *
        * ```ts
        * this.chart.plotAreaBackground="#F0F8FF";
        * ```
        *
        * ```ts
        *  <IgrDataChart
        * 	dataSource={this.state.dataSource}
        * 	plotAreaBackground="#F0F8FF" >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return brushToString(this.i.vk);
        },
        set: function (v) {
            this.i.vk = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "shouldMatchZOrderToSeriesOrder", {
        /**
         * Gets or sets whether to match the z order of the series to their positional order in the series collection, regardless of mutation.
        */
        get: function () {
            return this.i.f5;
        },
        set: function (v) {
            this.i.f5 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "defaultInteraction", {
        /**
         * Gets or sets the DefaultInteraction property.
         * The default interaction state defines the SeriesViewer's response to mouse events.
        *
        * The default interaction state defines the chart's response to mouse events.
        *
        * ```ts
        * this.chart.defaultInteraction = InteractionState.DragPan;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	defaultInteraction="dragPan" >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrAreaSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value"
        *     brush="Gray"
        *     outline="Black" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.cw;
        },
        set: function (v) {
            this.i.cw = ensureEnum(InteractionState_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "interactionOverride", {
        /**
         * Gets or sets the InteractionOverride property.
         * The interaction override overrides the default interaction and modifiers keys and defines the SeriesViewer's response to mouse events.
        */
        get: function () {
            return this.i.cx;
        },
        set: function (v) {
            this.i.cx = ensureEnum(InteractionState_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "rightButtonDefaultInteraction", {
        /**
         * Gets or sets the RightButtonDefaultInteraction property.
         * The default interaction state defines the SeriesViewer's response to right button mouse events.
        */
        get: function () {
            return this.i.cz;
        },
        set: function (v) {
            this.i.cz = ensureEnum(InteractionState_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "dragModifier", {
        /**
         * Gets or sets the current SeriesViewer's DragModifier property.
        *
        * `DragModifier` is useful to enable zooming when zooming is not the default interaction.
        *
        * ```ts
        * this.chart.dragModifier = ModifierKeys.Alt;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	isHorizontalZoomEnabled=true
        *     defaultInteraction="dragPan"
        *     dragModifier="alt" >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.u1;
        },
        set: function (v) {
            this.i.u1 = ensureEnum(ModifierKeys_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "panModifier", {
        /**
         * Gets or sets the current SeriesViewer's PanModifier property.
        *
        * `PanModifier` is useful to enable panning when panning is not the default interaction.
        *
        * ```ts
        * this.chart.panModifier = ModifierKeys.Alt;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        * 	dataSource={this.state.dataSource}
        * 	isHorizontalZoomEnabled=true
        * 	defaultInteraction="dragZoom"
        * 	panModifier="alt" >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.u2;
        },
        set: function (v) {
            this.i.u2 = ensureEnum(ModifierKeys_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "selectionModifier", {
        /**
         * Gets or sets the current SeriesViewer's SelectionModifier property.
        */
        get: function () {
            return this.i.u3;
        },
        set: function (v) {
            this.i.u3 = ensureEnum(ModifierKeys_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "previewRect", {
        /**
         * Gets or sets the preview rectangle.
         * The preview rectangle may be set to Rect.Empty, in which case the visible preview
         * strokePath is hidden.
        *
        * `PreviewRect` can be used to highlight an area of importance.
        *
        * ```ts
        * this.chart.previewRect = {left:0,top:0,height:.5,width:.5};
        * ```
        */
        get: function () {
            return fromRect(this.i.we);
        },
        set: function (v) {
            this.i.we = toRect(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "highlightedValuesDisplayMode", {
        /**
         * Gets or sets whether and how to display highlighted values for the series by default. Note, this is distinct from the highlighting feature that indicates what is closest or under the mouse.
        */
        get: function () {
            return this.i.bt;
        },
        set: function (v) {
            this.i.bt = ensureEnum(SeriesHighlightedValuesDisplayMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "selectedSeriesItems", {
        /**
         * Gets the currently selected data items.  Adding or removing data items from this collection will
         * select or deselect the visuals associated with those items.
        */
        get: function () {
            if (this._selectedSeriesItems === null) {
                var coll = new IgrChartSelectedItemCollection();
                var innerColl = this.i.selectedSeriesItems;
                if (!innerColl) {
                    innerColl = new ChartSelectedItemCollection_internal();
                }
                this._selectedSeriesItems = coll._fromInner(innerColl);
                this.i.selectedSeriesItems = innerColl;
            }
            return this._selectedSeriesItems;
        },
        set: function (v) {
            if (this._selectedSeriesItems !== null) {
                this._selectedSeriesItems._setSyncTarget(null);
                this._selectedSeriesItems = null;
            }
            var coll = new IgrChartSelectedItemCollection();
            this._selectedSeriesItems = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(ChartSelection.$type);
            var innerColl = this.i.selectedSeriesItems;
            if (!innerColl) {
                innerColl = new ChartSelectedItemCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._selectedSeriesItems._setSyncTarget(syncColl);
            this.i.selectedSeriesItems = innerColl;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "focusedSeriesItems", {
        /**
         * Gets the currently focused data items.  Adding or removing data items from this collection will
         * focus or blur the visuals associated with those items.
        */
        get: function () {
            if (this._focusedSeriesItems === null) {
                var coll = new IgrChartSelectedItemCollection();
                var innerColl = this.i.focusedSeriesItems;
                if (!innerColl) {
                    innerColl = new ChartSelectedItemCollection_internal();
                }
                this._focusedSeriesItems = coll._fromInner(innerColl);
                this.i.focusedSeriesItems = innerColl;
            }
            return this._focusedSeriesItems;
        },
        set: function (v) {
            if (this._focusedSeriesItems !== null) {
                this._focusedSeriesItems._setSyncTarget(null);
                this._focusedSeriesItems = null;
            }
            var coll = new IgrChartSelectedItemCollection();
            this._focusedSeriesItems = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(ChartSelection.$type);
            var innerColl = this.i.focusedSeriesItems;
            if (!innerColl) {
                innerColl = new ChartSelectedItemCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._focusedSeriesItems._setSyncTarget(syncColl);
            this.i.focusedSeriesItems = innerColl;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "windowPositionHorizontal", {
        /**
         * A number between 0 and 1 determining the position of the horizontal scroll.
         * This property is effectively a shortcut to the X position of the WindowRect property.
        *
        * ```ts
        * this.chart.windowPositionHorizontal = .2;
        * ```
        */
        get: function () {
            return this.i.iq;
        },
        set: function (v) {
            this.i.iq = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "windowPositionVertical", {
        /**
         * A number between 0 and 1 determining the position of the vertical scroll.
         * This property is effectively a shortcut to the Y position of the WindowRect property.
        *
        * ```ts
        * this.chart.windowPositionVertical = .2;
        * ```
        */
        get: function () {
            return this.i.ir;
        },
        set: function (v) {
            this.i.ir = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "selectionBrush", {
        /**
         * Gets or sets the selection brush to use for the series.
        */
        get: function () {
            return brushToString(this.i.vo);
        },
        set: function (v) {
            this.i.vo = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "focusBrush", {
        /**
         * Gets or sets the Focus brush to use for the series.
        */
        get: function () {
            return brushToString(this.i.u7);
        },
        set: function (v) {
            this.i.u7 = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "chartTitle", {
        /**
         * Text to display above the plot area.
        *
        * The `Title` property is used to set the title to display for the component.
        *
        * ```ts
        * this.chart.title="Items Sold";
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	title="Items Sold" >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.l3;
        },
        set: function (v) {
            this.i.l3 = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "titleHorizontalAlignment", {
        /**
         * Gets or sets the TitleHorizontalAlignment property.
        *
        * The `subtitleHorizontalAlignment` property is used to set horizontal alignment for the subtitle.
        *
        * ```ts
        * this.chart.titleHorizontalAlignment="left";
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	titleHorizontalAlignment="left"
        *     subtitleHorizontalAlignment="left" >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.u0;
        },
        set: function (v) {
            this.i.u0 = ensureEnum(HorizontalAlignment_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "subtitleHorizontalAlignment", {
        /**
         * Gets or sets the SubtitleHorizontalAlignment property.
        *
        * The `subtitleBottomMargin` property is used to set the horizontal alignment for the subtitle.
        *
        * ```ts
        * this.chart.subtitleHorizontalAlignment="left";
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	subtitleHorizontalAlignment="left"	>
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.uz;
        },
        set: function (v) {
            this.i.uz = ensureEnum(HorizontalAlignment_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "titleTextStyle", {
        /**
         * Font settings for the title.
        */
        get: function () {
            return this.i.ma;
        },
        set: function (v) {
            this.i.ma = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "subtitleTextStyle", {
        /**
         * Font settings for the subtitle.
        */
        get: function () {
            return this.i.ly;
        },
        set: function (v) {
            this.i.ly = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "titleTextColor", {
        /**
         * Gets or sets the Color used for the Title Text.
        *
        * The `titleTextColor` property is used to set the title brush.
        *
        * ```ts
        * this.chart.titleTextColor="blue" ;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	titleTextColor="blue" >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return brushToString(this.i.vq);
        },
        set: function (v) {
            this.i.vq = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "subtitleTextColor", {
        /**
         * Gets or sets the Color used for the Subtitle Text.
        *
        * The `subtitleTextColor` property is used to set the color for the subtitle.
        *
        * ```ts
        * this.chart.subtitleTextColor="blue";
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	title="Date of Sale: 1/1/2018"
        *     subtitleTextColor="blue">
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return brushToString(this.i.vp);
        },
        set: function (v) {
            this.i.vp = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "titleTopMargin", {
        /**
         * Margin applied above the title.
        *
        * The `titleTopMargin` property is used to set the top margin for the title.
        *
        * ```ts
        * this.chart.titleTopMargin=10;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	titleLeftMarging= {10}
        *     titleTopMargin= {10 }
        *     titleRightMargin= {10}
        *     titleBottomMargin= {20}>
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.ig;
        },
        set: function (v) {
            this.i.ig = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "titleLeftMargin", {
        /**
         * Margin applied to the left of the title.
        *
        * The `titleLeftMarging` property is used to set the left margin for the title.
        *
        * ```ts
        * this.chart.titleLeftMarging=10;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	title="Items Sold"
        *     titleLeftMarging={10}
        *     titleTopMargin={10}
        *     titleRightMargin={10}
        *     titleBottomMargin= {20} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.id;
        },
        set: function (v) {
            this.i.id = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "titleRightMargin", {
        /**
         * Margin applied to the right of the title.
        *
        * The `titleRightMargin` property is used to set the right margin for the title.
        *
        * ```ts
        * this.chart.titleRightMargin=10;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	titleLeftMarging= {10}
        *     titleTopMargin= {10 }
        *     titleRightMargin= {10}
        *     titleBottomMargin= {20} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.ie;
        },
        set: function (v) {
            this.i.ie = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "titleBottomMargin", {
        /**
         * Margin applied below the title.
        *
        * The `titleBottomMargin` property is used to set the bottom margin for the title.
        *
        * ```ts
        * this.chart.titleBottomMargin=20;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	title="Items Sold"
        * 	titleLeftMarging= {10}
        *     titleTopMargin= {10 }
        *     titleRightMargin= {10}
        *     titleBottomMargin= {20}>
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.ic;
        },
        set: function (v) {
            this.i.ic = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "subtitleTopMargin", {
        /**
         * Margin applied above the subtitle.
        *
        * The `subtitleTopMargin` property is used to set the top margin for the subtitle.
        *
        * ```ts
        * this.chart.subtitleTopMargin =10;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	subtitleLeftMarging= {10}
        *     subtitleTopMargin={10}
        *     subtitleRightMargin={10}
        *     subtitleBottomMargin={20}>
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.ib;
        },
        set: function (v) {
            this.i.ib = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "subtitleLeftMargin", {
        /**
         * Margin applied to the left of the subtitle.
        *
        * The `subtitleLeftMarging` property is used to set the left margin for the subtitle.
        *
        * ```ts
        * this.chart.subtitleLeftMarging=10;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	subtitleLeftMarging= {10}
        *     subtitleTopMargin={10}
        *     subtitleRightMargin={10}
        *     subtitleBottomMargin={20}>
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.h9;
        },
        set: function (v) {
            this.i.h9 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "subtitleRightMargin", {
        /**
         * Margin applied to the right of the subtitle.
        *
        * The `subTitleRightMargin` property is used to set the right margin for the subtitle.
        *
        * ```ts
        * this.chart.subtitleLeftMarging=10;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	subtitleLeftMarging= {10}
        *     subtitleTopMargin= {10}
        *     subtitleRightMargin={10}
        *     subtitleBottomMargin={20} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.ia;
        },
        set: function (v) {
            this.i.ia = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "subtitleBottomMargin", {
        /**
         * Margin applied below the subtitle.
        *
        * The `subtitleBottomMargin` property is used to set the bottom margin for the subtitle.
        *
        * ```ts
        * this.chart.subtitleBottomMargin = 20;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	 subtitleBottomMargin={20} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.h8;
        },
        set: function (v) {
            this.i.h8 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "subtitle", {
        /**
         * Gets or sets the Subtitle property.
         * The Title property defines the Subtitle of the chart
        *
        * The `Subtitle` property is used to display subtitle for the component.
        *
        * ```ts
        * this.chart.subtitle="Date of Sale: 1/1/2018";
        * ```
        *
        * ```ts
        *  <IgrDataChart
        * 	dataSource={this.state.dataSource}
        * 	subtitle="Date of Sale: 1/1/2018" >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.lr;
        },
        set: function (v) {
            this.i.lr = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "topMargin", {
        /**
         * Sets or gets the top margin to use around the chart content.
        *
        * The `topMargin` property is used to set the top margin.
        *
        * ```ts
        * this.chart.topMargin=30;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	leftMargin= {20}
        *     rightMargin={20}
        *     topMargin={30}
        *     bottomMargin= {20} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.ih;
        },
        set: function (v) {
            this.i.ih = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "leftMargin", {
        /**
         * Sets or gets the left margin to use around the chart content.
        *
        * The `LeftMargin` property is used to set the left margin.
        *
        * ```ts
        *  this.chart.leftMargin = 20;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	 leftMargin={20} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value"
        * />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.h2;
        },
        set: function (v) {
            this.i.h2 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "rightMargin", {
        /**
         * Sets or gets the right margin to use around the chart content.
        *
        * The `RightMargin` property is used to set the left margin around the chart content in the canvas.
        *
        * ```ts
        * this.chart.rightMargin=20;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	leftMargin= {20}
        * 	rightMargin={20}
        * 	topMargin={30}
        * 	bottomMargin={20} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.h7;
        },
        set: function (v) {
            this.i.h7 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "bottomMargin", {
        /**
         * Sets or gets the bottom margin to use around the chart content.
        *
        * The `bottomMargin` property is used to gets/set the bottom margin around the chart content in the canvas.
        *
        * ```ts
        * this.chart.bottomMargin = 50;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	 bottomMargin={50} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrColumnSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value"
        *     brush="Gray"
        *     outline="Black" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.hp;
        },
        set: function (v) {
            this.i.hp = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "autoMarginWidth", {
        /**
         * Sets or gets the automatic width to add when automatically adding margins to the chart.
        *
        * The `autoMarginWidth` propertry is used find the automatic width to add when automatically adding margins to the chart.
        *
        * ```ts
        * this.chart.autoMarginWidth = 50;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	autoMarginWidth={50} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrColumnSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value"
        *     brush="Gray"
        *     outline="Black" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.ho;
        },
        set: function (v) {
            this.i.ho = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "autoMarginHeight", {
        /**
         * Sets or gets the automatic height to add when automatically adding margins to the chart.
        *
        * ```ts
        *  this.chart.autoMarginHeight = 50;
        * ```
        *
        * ```ts
        *  <IgrDataChart
        * 	autoMarginHeight={50}
        *     dataSource={this.state.dataSource} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrAreaSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value"
        *     brush="Gray"
        *     outline="Black" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this.i.hn;
        },
        set: function (v) {
            this.i.hn = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "isAntiAliasingEnabledDuringInteraction", {
        /**
         * Gets or sets the IsAntiAliasingEnabledDuringInteraction property.
        */
        get: function () {
            return this.i.e3;
        },
        set: function (v) {
            this.i.e3 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "pixelScalingRatio", {
        /**
         * Gets or sets the scaling value used to affect the pixel density of the control.
         * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
         * to appear blurry.
        */
        get: function () {
            return this.i.h3;
        },
        set: function (v) {
            this.i.h3 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "interactionPixelScalingRatio", {
        /**
         * Method of pixel scaling to use during end user interaction with the chart.
        */
        get: function () {
            return this.i.h1;
        },
        set: function (v) {
            this.i.h1 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "actualPixelScalingRatio", {
        /**
         * Gets the resolved pixel scaling ratio.
        *
        * A higher scaling ratio will produce crisper visuals at the expense of memory. Lower values will cause the control to appear blurry
        *
        * ```ts
        * let actualPixelScalingRatio: number = this.chart.actualPixelScalingRatio;
        * ```
        */
        get: function () {
            return this.i.hi;
        },
        set: function (v) {
            this.i.hi = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "actualInteractionPixelScalingRatio", {
        /**
         * Resolved method of pixel scaling to use during end user interaction with the chart.
        *
        * ```ts
        * let actualInteractionPixelScalingRatio: number = this.chart.actualInteractionPixelScalingRatio;
        * ```
        */
        get: function () {
            return this.i.hh;
        },
        set: function (v) {
            this.i.hh = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "shouldPanOnMaximumZoom", {
        /**
         * Gets or sets a whether the chart should pan its content when zooming in passed max zoom level.
        */
        get: function () {
            return this.i.f6;
        },
        set: function (v) {
            this.i.f6 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "actualWindowRect", {
        /**
         * Gets the actual value of the window rectangle, which represents the current zoom level.
        */
        get: function () {
            return fromRect(this.i.v4);
        },
        set: function (v) {
            this.i.v4 = toRect(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "actualWindowPositionHorizontal", {
        /**
         * A number between 0 and 1 determining the position of the horizontal scroll.
         * This property is effectively a shortcut to the Left of the ActualWindowRect property.
        *
        * ```ts
        * let actualWindowPositionHorizontal: number = this.chart.actualWindowPositionHorizontal;
        * ```
        */
        get: function () {
            return this.i.hj;
        },
        set: function (v) {
            this.i.hj = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "actualWindowPositionVertical", {
        /**
         * A number between 0 and 1 determining the position of the vertical scroll.
         * This property is effectively a shortcut to the Top of the ActualWindowRect property.
        *
        * ```ts
        * let actualWindowPositionVertical: number = this.chart.actualWindowPositionVertical;
        * ``
        * ```
        */
        get: function () {
            return this.i.hk;
        },
        set: function (v) {
            this.i.hk = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "previewPathStroke", {
        /**
         * Gets or sets the brush used as the border for the current SeriesViewer object's preview path.
         * Null is treated as auto.
        */
        get: function () {
            return brushToString(this.i.vn);
        },
        set: function (v) {
            this.i.vn = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "previewPathFill", {
        /**
         * Gets or sets the brush used as the fill for the current SeriesViewer object's preview path.
         * Null is treated as auto.
        */
        get: function () {
            return brushToString(this.i.vm);
        },
        set: function (v) {
            this.i.vm = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "previewPathOpacity", {
        /**
         * Gets or sets the opacity used as for the fill for the current SeriesViewer object's preview path.
         * Null is treated as auto.
        */
        get: function () {
            return this.i.h4;
        },
        set: function (v) {
            this.i.h4 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "isDetached", {
        get: function () {
            return this.i.isDetached;
        },
        enumerable: false,
        configurable: true
    });
    IgrSeriesViewer.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.series != null && this.series.findByName && this.series.findByName(name)) {
            return this.series.findByName(name);
        }
        if (this.legend && this.legend.name && this.legend.name == name) {
            return this.legend;
        }
        if (this.selectedSeriesItems != null && this.selectedSeriesItems.findByName && this.selectedSeriesItems.findByName(name)) {
            return this.selectedSeriesItems.findByName(name);
        }
        if (this.focusedSeriesItems != null && this.focusedSeriesItems.findByName && this.focusedSeriesItems.findByName(name)) {
            return this.focusedSeriesItems.findByName(name);
        }
        return null;
    };
    Object.defineProperty(IgrSeriesViewer.prototype, "hasUserValues", {
        get: function () {
            return this._hasUserValues;
        },
        enumerable: false,
        configurable: true
    });
    IgrSeriesViewer.prototype.__m = function (propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    };
    IgrSeriesViewer.prototype._styling = function (container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        var genericPrefix = "";
        var typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("SeriesViewer");
        var additionalPrefixes = [];
        var prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        var b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            var basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            var parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            var parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    };
    IgrSeriesViewer.prototype.getDesiredToolbarActions = function () {
        var iv = this.i.getDesiredToolbarActions();
        return (iv);
    };
    IgrSeriesViewer.prototype.addCommandAvailabilityListener = function (listener) {
        this.i.addCommandAvailabilityListener(listener);
    };
    IgrSeriesViewer.prototype.removeCommandAvailabilityListener = function (listener) {
        this.i.removeCommandAvailabilityListener(listener);
    };
    IgrSeriesViewer.prototype.addCommandStateChangedListener = function (listener) {
        this.i.addCommandStateChangedListener(listener);
    };
    IgrSeriesViewer.prototype.removeCommandStateChangedListener = function (listener) {
        this.i.removeCommandStateChangedListener(listener);
    };
    /**
     * Hides the active main tooltip, if displayed.
    
    *
    * ```ts
    * this.chart.hideToolTip();
    * ```
    */
    IgrSeriesViewer.prototype.hideToolTip = function () {
        this.i.nx();
    };
    IgrSeriesViewer.prototype.simulateHover = function (point) {
        this.i.rg(toPoint(point));
    };
    IgrSeriesViewer.prototype.queueForAnimationIdle = function (action, version) {
        this.i.queueForAnimationIdle(action, version);
    };
    IgrSeriesViewer.prototype.getAnimationIdleVersionNumber = function () {
        var iv = this.i.getAnimationIdleVersionNumber();
        return (iv);
    };
    IgrSeriesViewer.prototype.isAnimationActive = function () {
        var iv = this.i.isAnimationActive();
        return (iv);
    };
    /**
     * Manually starts a tiled zoom if one isn't already running.
    
    */
    IgrSeriesViewer.prototype.startTiledZoomingIfNecessary = function () {
        this.i.ri();
    };
    /**
     * Manually ends a tiled zoom if one is running.
    
    */
    IgrSeriesViewer.prototype.endTiledZoomingIfRunning = function () {
        this.i.nl();
    };
    /**
     * Clears the tile zoom tile cache so that new tiles will be generated. Only applies if the viewer is using a tile based zoom.
    
    */
    IgrSeriesViewer.prototype.clearTileZoomCache = function () {
        this.i.nf();
    };
    /**
     * Performs zoom in action on the chart by specified percentage of current window rect
    
    */
    IgrSeriesViewer.prototype.zoomIn = function (percentage) {
        this.i.rz(percentage);
    };
    /**
     * Performs zoom out action on the chart by specified percentage of current window rect
    
    */
    IgrSeriesViewer.prototype.zoomOut = function (percentage) {
        this.i.r0(percentage);
    };
    /**
     * Resets the zoom level to default.
    
    */
    IgrSeriesViewer.prototype.resetZoom = function () {
        this.i.q8();
    };
    IgrSeriesViewer.prototype.cancelManipulation = function () {
        this.i.na();
    };
    /**
     * Gets actual window scale for horizontal dimension of the control
    
    *
    * ```ts
    * let actualWindowScaleHorizontal: number = chart.getActualWindowScaleHorizontal();
    * ```
    */
    IgrSeriesViewer.prototype.getActualWindowScaleHorizontal = function () {
        var iv = this.i.hs();
        return (iv);
    };
    /**
     * Gets actual window scale for vertical dimension of the control
    
    *
    * ```ts
    * let actualWindowScaleVertical: number = chart.getActualWindowScaleVertical();
    * ```
    */
    IgrSeriesViewer.prototype.getActualWindowScaleVertical = function () {
        var iv = this.i.ht();
        return (iv);
    };
    /**
     * Notification from the containing object that the container has been resized.
    
    */
    IgrSeriesViewer.prototype.notifyContainerResized = function () {
        this.i.notifyContainerResized();
    };
    /**
     * Attaches the given series to the chart.
    
    * @param s  * The series to attach to the chart.
    */
    IgrSeriesViewer.prototype.attachSeries = function (s) {
        this.i.m8((s == null ? null : s.i));
    };
    IgrSeriesViewer.prototype.notifySetItem = function (source_, index, oldItem, newItem) {
        this.i.og(source_, index, oldItem, newItem);
    };
    /**
     * Used to manually notify the SeriesViewer that the data source has reset or cleared its items.  Invoking this method is only necessary if that datasource is not observable.
    
    * @param source_  * The data source which was cleared or reset.
    *
    * ```ts
    * this.chart.notifyClearItems(this.data);
    * ```
    */
    IgrSeriesViewer.prototype.notifyClearItems = function (source_) {
        this.i.n7(source_);
    };
    IgrSeriesViewer.prototype.notifyInsertItem = function (source_, index, newItem) {
        this.i.oe(source_, index, newItem);
    };
    IgrSeriesViewer.prototype.notifyRemoveItem = function (source_, index, oldItem) {
        this.i.of(source_, index, oldItem);
    };
    /**
     * Release any resources held by the control.
    
    */
    IgrSeriesViewer.prototype.destroy = function () {
        this.i.destroy();
    };
    IgrSeriesViewer.prototype.getCurrentActualWindowRect = function () {
        var iv = this.i.wa();
        return fromRect(iv);
    };
    IgrSeriesViewer.prototype.renderToImage = function (width, height) {
        var iv = this.i.jb(width, height);
        return (iv);
    };
    /**
     * Use to force the SeriesViewer to finish any deferred work before printing or evaluating its visual.
     * This should only be called if the visual of the SeriesViewer needs to be synchronously saved or evaluated.
     * Calling this method too often will hinder the performance of the SeriesViewer.
    
    *
    * ```ts
    * this.chart.flush();
    * ```
    */
    IgrSeriesViewer.prototype.flush = function () {
        this.i.flush();
    };
    IgrSeriesViewer.prototype.captureImage = function (settings) {
        this.i.nb((settings == null ? null : settings.i));
    };
    Object.defineProperty(IgrSeriesViewer.prototype, "plotAreaMouseLeftButtonDown", {
        /**
         * Occurs when the left mouse button is pressed while the mouse pointer is over the plot area.
        */
        get: function () {
            return this._plotAreaMouseLeftButtonDown;
        },
        set: function (ev) {
            var _this = this;
            if (this._plotAreaMouseLeftButtonDown_wrapped !== null) {
                this.i.plotAreaMouseLeftButtonDown = delegateRemove(this.i.plotAreaMouseLeftButtonDown, this._plotAreaMouseLeftButtonDown_wrapped);
                this._plotAreaMouseLeftButtonDown_wrapped = null;
                this._plotAreaMouseLeftButtonDown = null;
            }
            this._plotAreaMouseLeftButtonDown = ev;
            this._plotAreaMouseLeftButtonDown_wrapped = function (o, e) {
                var outerArgs = new IgrPlotAreaMouseButtonEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforePlotAreaMouseLeftButtonDown) {
                    _this.beforePlotAreaMouseLeftButtonDown(_this, outerArgs);
                }
                if (_this._plotAreaMouseLeftButtonDown) {
                    _this._plotAreaMouseLeftButtonDown(_this, outerArgs);
                }
            };
            this.i.plotAreaMouseLeftButtonDown = delegateCombine(this.i.plotAreaMouseLeftButtonDown, this._plotAreaMouseLeftButtonDown_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "plotAreaMouseLeftButtonUp", {
        /**
         * Occurs when the left mouse button is pressed while the mouse pointer is over the plot area.
        */
        get: function () {
            return this._plotAreaMouseLeftButtonUp;
        },
        set: function (ev) {
            var _this = this;
            if (this._plotAreaMouseLeftButtonUp_wrapped !== null) {
                this.i.plotAreaMouseLeftButtonUp = delegateRemove(this.i.plotAreaMouseLeftButtonUp, this._plotAreaMouseLeftButtonUp_wrapped);
                this._plotAreaMouseLeftButtonUp_wrapped = null;
                this._plotAreaMouseLeftButtonUp = null;
            }
            this._plotAreaMouseLeftButtonUp = ev;
            this._plotAreaMouseLeftButtonUp_wrapped = function (o, e) {
                var outerArgs = new IgrPlotAreaMouseButtonEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforePlotAreaMouseLeftButtonUp) {
                    _this.beforePlotAreaMouseLeftButtonUp(_this, outerArgs);
                }
                if (_this._plotAreaMouseLeftButtonUp) {
                    _this._plotAreaMouseLeftButtonUp(_this, outerArgs);
                }
            };
            this.i.plotAreaMouseLeftButtonUp = delegateCombine(this.i.plotAreaMouseLeftButtonUp, this._plotAreaMouseLeftButtonUp_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "plotAreaClicked", {
        /**
         * Occurs when the left mouse button is clicked while the mouse pointer is over the plot area.
        */
        get: function () {
            return this._plotAreaClicked;
        },
        set: function (ev) {
            var _this = this;
            if (this._plotAreaClicked_wrapped !== null) {
                this.i.plotAreaClicked = delegateRemove(this.i.plotAreaClicked, this._plotAreaClicked_wrapped);
                this._plotAreaClicked_wrapped = null;
                this._plotAreaClicked = null;
            }
            this._plotAreaClicked = ev;
            this._plotAreaClicked_wrapped = function (o, e) {
                var outerArgs = new IgrPlotAreaMouseButtonEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforePlotAreaClicked) {
                    _this.beforePlotAreaClicked(_this, outerArgs);
                }
                if (_this._plotAreaClicked) {
                    _this._plotAreaClicked(_this, outerArgs);
                }
            };
            this.i.plotAreaClicked = delegateCombine(this.i.plotAreaClicked, this._plotAreaClicked_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "plotAreaMouseEnter", {
        /**
         * Occurs when the pointer enters the plot area.
        */
        get: function () {
            return this._plotAreaMouseEnter;
        },
        set: function (ev) {
            var _this = this;
            if (this._plotAreaMouseEnter_wrapped !== null) {
                this.i.plotAreaMouseEnter = delegateRemove(this.i.plotAreaMouseEnter, this._plotAreaMouseEnter_wrapped);
                this._plotAreaMouseEnter_wrapped = null;
                this._plotAreaMouseEnter = null;
            }
            this._plotAreaMouseEnter = ev;
            this._plotAreaMouseEnter_wrapped = function (o, e) {
                var outerArgs = new IgrPlotAreaMouseEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforePlotAreaMouseEnter) {
                    _this.beforePlotAreaMouseEnter(_this, outerArgs);
                }
                if (_this._plotAreaMouseEnter) {
                    _this._plotAreaMouseEnter(_this, outerArgs);
                }
            };
            this.i.plotAreaMouseEnter = delegateCombine(this.i.plotAreaMouseEnter, this._plotAreaMouseEnter_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "plotAreaMouseLeave", {
        /**
         * Occurs when the pointer exits the plot area.
        */
        get: function () {
            return this._plotAreaMouseLeave;
        },
        set: function (ev) {
            var _this = this;
            if (this._plotAreaMouseLeave_wrapped !== null) {
                this.i.plotAreaMouseLeave = delegateRemove(this.i.plotAreaMouseLeave, this._plotAreaMouseLeave_wrapped);
                this._plotAreaMouseLeave_wrapped = null;
                this._plotAreaMouseLeave = null;
            }
            this._plotAreaMouseLeave = ev;
            this._plotAreaMouseLeave_wrapped = function (o, e) {
                var outerArgs = new IgrPlotAreaMouseEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforePlotAreaMouseLeave) {
                    _this.beforePlotAreaMouseLeave(_this, outerArgs);
                }
                if (_this._plotAreaMouseLeave) {
                    _this._plotAreaMouseLeave(_this, outerArgs);
                }
            };
            this.i.plotAreaMouseLeave = delegateCombine(this.i.plotAreaMouseLeave, this._plotAreaMouseLeave_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "plotAreaMouseOver", {
        /**
         * Occurs when the pointer if over the plot area.
        */
        get: function () {
            return this._plotAreaMouseOver;
        },
        set: function (ev) {
            var _this = this;
            if (this._plotAreaMouseOver_wrapped !== null) {
                this.i.plotAreaMouseOver = delegateRemove(this.i.plotAreaMouseOver, this._plotAreaMouseOver_wrapped);
                this._plotAreaMouseOver_wrapped = null;
                this._plotAreaMouseOver = null;
            }
            this._plotAreaMouseOver = ev;
            this._plotAreaMouseOver_wrapped = function (o, e) {
                var outerArgs = new IgrPlotAreaMouseEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforePlotAreaMouseOver) {
                    _this.beforePlotAreaMouseOver(_this, outerArgs);
                }
                if (_this._plotAreaMouseOver) {
                    _this._plotAreaMouseOver(_this, outerArgs);
                }
            };
            this.i.plotAreaMouseOver = delegateCombine(this.i.plotAreaMouseOver, this._plotAreaMouseOver_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "seriesCursorMouseMove", {
        /**
         * Occurs when the cursors are moved over a series in this SeriesViewer.
        */
        get: function () {
            return this._seriesCursorMouseMove;
        },
        set: function (ev) {
            var _this = this;
            if (this._seriesCursorMouseMove_wrapped !== null) {
                this.i.seriesCursorMouseMove = delegateRemove(this.i.seriesCursorMouseMove, this._seriesCursorMouseMove_wrapped);
                this._seriesCursorMouseMove_wrapped = null;
                this._seriesCursorMouseMove = null;
            }
            this._seriesCursorMouseMove = ev;
            this._seriesCursorMouseMove_wrapped = function (o, e) {
                var outerArgs = new IgrChartCursorEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeSeriesCursorMouseMove) {
                    _this.beforeSeriesCursorMouseMove(_this, outerArgs);
                }
                if (_this._seriesCursorMouseMove) {
                    _this._seriesCursorMouseMove(_this, outerArgs);
                }
            };
            this.i.seriesCursorMouseMove = delegateCombine(this.i.seriesCursorMouseMove, this._seriesCursorMouseMove_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "seriesMouseLeftButtonDown", {
        /**
         * Occurs when the left mouse button is pressed while the mouse pointer is over a Series.
        *
        * The `SeriesMouseLeftButtonDown` event occurs when the left mouse button is pressed while the mouse pointer is over an element of this chart.
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        * public onSeriesMouseLeftButtonDown(event: {sender: any, args: DataChartMouseButtonEventArgs})
        * {
        *     var item = event.args.item;
        * }
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	seriesMouseLeftButtonDown= {this.onSeriesMouseLeftButtonDown}>
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this._seriesMouseLeftButtonDown;
        },
        set: function (ev) {
            var _this = this;
            if (this._seriesMouseLeftButtonDown_wrapped !== null) {
                this.i.seriesMouseLeftButtonDown = delegateRemove(this.i.seriesMouseLeftButtonDown, this._seriesMouseLeftButtonDown_wrapped);
                this._seriesMouseLeftButtonDown_wrapped = null;
                this._seriesMouseLeftButtonDown = null;
            }
            this._seriesMouseLeftButtonDown = ev;
            this._seriesMouseLeftButtonDown_wrapped = function (o, e) {
                var outerArgs = new IgrDataChartMouseButtonEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeSeriesMouseLeftButtonDown) {
                    _this.beforeSeriesMouseLeftButtonDown(_this, outerArgs);
                }
                if (_this._seriesMouseLeftButtonDown) {
                    _this._seriesMouseLeftButtonDown(_this, outerArgs);
                }
            };
            this.i.seriesMouseLeftButtonDown = delegateCombine(this.i.seriesMouseLeftButtonDown, this._seriesMouseLeftButtonDown_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "seriesMouseLeftButtonUp", {
        /**
         * Occurs when the left mouse button is released while the mouse pointer is over a Series.
        *
        * The `SeriesMouseLeftButtonUp` event occurs when the left mouse button is released while the mouse pointer is over an element of this chart.
        *
        * &lt;-- position: content member-->
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        * public onSeriesMouseLeftButtonUp(event: {sender: any, args: DataChartMouseButtonEventArgs})
        * {
        *     var item = event.args.item;
        * }
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	seriesMouseLeftButtonUp ={this.onSeriesMouseLeftButtonUp}>
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        *
        * ```ts
        * public onSeriesMouseLeftButtonUp =(event: any, args: DataChartMouseButtonEventArgs )
        * {
        *
        * }
        * ```
        */
        get: function () {
            return this._seriesMouseLeftButtonUp;
        },
        set: function (ev) {
            var _this = this;
            if (this._seriesMouseLeftButtonUp_wrapped !== null) {
                this.i.seriesMouseLeftButtonUp = delegateRemove(this.i.seriesMouseLeftButtonUp, this._seriesMouseLeftButtonUp_wrapped);
                this._seriesMouseLeftButtonUp_wrapped = null;
                this._seriesMouseLeftButtonUp = null;
            }
            this._seriesMouseLeftButtonUp = ev;
            this._seriesMouseLeftButtonUp_wrapped = function (o, e) {
                var outerArgs = new IgrDataChartMouseButtonEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeSeriesMouseLeftButtonUp) {
                    _this.beforeSeriesMouseLeftButtonUp(_this, outerArgs);
                }
                if (_this._seriesMouseLeftButtonUp) {
                    _this._seriesMouseLeftButtonUp(_this, outerArgs);
                }
            };
            this.i.seriesMouseLeftButtonUp = delegateCombine(this.i.seriesMouseLeftButtonUp, this._seriesMouseLeftButtonUp_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "seriesClick", {
        /**
         * Occurs when the left mouse button is released while the mouse pointer is over a Series.
        */
        get: function () {
            return this._seriesClick;
        },
        set: function (ev) {
            var _this = this;
            if (this._seriesClick_wrapped !== null) {
                this.i.seriesClick = delegateRemove(this.i.seriesClick, this._seriesClick_wrapped);
                this._seriesClick_wrapped = null;
                this._seriesClick = null;
            }
            this._seriesClick = ev;
            this._seriesClick_wrapped = function (o, e) {
                var outerArgs = new IgrDataChartMouseButtonEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeSeriesClick) {
                    _this.beforeSeriesClick(_this, outerArgs);
                }
                if (_this._seriesClick) {
                    _this._seriesClick(_this, outerArgs);
                }
            };
            this.i.seriesClick = delegateCombine(this.i.seriesClick, this._seriesClick_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "seriesMouseMove", {
        /**
         * Occurs when the mouse pointer moves while over a Series.
        *
        * The `SeriesMouseMove` event occurs when the left mouse pointer moves while over an element of this chart.
        *
        * &lt;-- position: content member-->
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        * public onSeriesMouseMove(event: {sender: any, args: ChartMouseEventArgs})
        * {
        *     var item = event.args.item;
        * }
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	seriesMouseMove={this.onSeriesMouseMove} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        *
        * ```ts
        * public onSeriesMouseMove =( event: any, args: ChartMouseEventArgs) => {
        *                  }
        * ```
        */
        get: function () {
            return this._seriesMouseMove;
        },
        set: function (ev) {
            var _this = this;
            if (this._seriesMouseMove_wrapped !== null) {
                this.i.seriesMouseMove = delegateRemove(this.i.seriesMouseMove, this._seriesMouseMove_wrapped);
                this._seriesMouseMove_wrapped = null;
                this._seriesMouseMove = null;
            }
            this._seriesMouseMove = ev;
            this._seriesMouseMove_wrapped = function (o, e) {
                var outerArgs = new IgrChartMouseEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeSeriesMouseMove) {
                    _this.beforeSeriesMouseMove(_this, outerArgs);
                }
                if (_this._seriesMouseMove) {
                    _this._seriesMouseMove(_this, outerArgs);
                }
            };
            this.i.seriesMouseMove = delegateCombine(this.i.seriesMouseMove, this._seriesMouseMove_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "seriesMouseEnter", {
        /**
         * Occurs when the mouse pointer enters a Series.
        *
        * The `SeriesMouseEnter` event occurs when the left mouse pointer enters an element of this chart.
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        * public onSeriesMouseEnter(event: {sender: any, args: ChartMouseEventArgs})
        * {
        *     var item = event.args.item;
        * }
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	seriesMouseEnter={this.onSeriesMouseEnter}>
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this._seriesMouseEnter;
        },
        set: function (ev) {
            var _this = this;
            if (this._seriesMouseEnter_wrapped !== null) {
                this.i.seriesMouseEnter = delegateRemove(this.i.seriesMouseEnter, this._seriesMouseEnter_wrapped);
                this._seriesMouseEnter_wrapped = null;
                this._seriesMouseEnter = null;
            }
            this._seriesMouseEnter = ev;
            this._seriesMouseEnter_wrapped = function (o, e) {
                var outerArgs = new IgrChartMouseEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeSeriesMouseEnter) {
                    _this.beforeSeriesMouseEnter(_this, outerArgs);
                }
                if (_this._seriesMouseEnter) {
                    _this._seriesMouseEnter(_this, outerArgs);
                }
            };
            this.i.seriesMouseEnter = delegateCombine(this.i.seriesMouseEnter, this._seriesMouseEnter_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "seriesMouseLeave", {
        /**
         * Occurs when the mouse pointer leaves a Series.
        *
        * The `SeriesMouseLeave` event occurs when the left mouse pointer leaves an element of this chart.
        *
        * ```ts
        * public onSeriesMouseLeave(event: {sender: any, args: ChartMouseEventArgs})
        * {
        *     var item = event.args.item;
        * }
        * ```
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	 seriesMouseLeave ={this.onSeriesMouseLeave}>
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        */
        get: function () {
            return this._seriesMouseLeave;
        },
        set: function (ev) {
            var _this = this;
            if (this._seriesMouseLeave_wrapped !== null) {
                this.i.seriesMouseLeave = delegateRemove(this.i.seriesMouseLeave, this._seriesMouseLeave_wrapped);
                this._seriesMouseLeave_wrapped = null;
                this._seriesMouseLeave = null;
            }
            this._seriesMouseLeave = ev;
            this._seriesMouseLeave_wrapped = function (o, e) {
                var outerArgs = new IgrChartMouseEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeSeriesMouseLeave) {
                    _this.beforeSeriesMouseLeave(_this, outerArgs);
                }
                if (_this._seriesMouseLeave) {
                    _this._seriesMouseLeave(_this, outerArgs);
                }
            };
            this.i.seriesMouseLeave = delegateCombine(this.i.seriesMouseLeave, this._seriesMouseLeave_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "viewerManipulationStarting", {
        /**
         * Raised when an ongoing manipulation of the series viewer is started.
        */
        get: function () {
            return this._viewerManipulationStarting;
        },
        set: function (ev) {
            var _this = this;
            if (this._viewerManipulationStarting_wrapped !== null) {
                this.i.viewerManipulationStarting = delegateRemove(this.i.viewerManipulationStarting, this._viewerManipulationStarting_wrapped);
                this._viewerManipulationStarting_wrapped = null;
                this._viewerManipulationStarting = null;
            }
            this._viewerManipulationStarting = ev;
            this._viewerManipulationStarting_wrapped = function (o, e) {
                var outerArgs = new IgrSeriesViewerManipulationEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeViewerManipulationStarting) {
                    _this.beforeViewerManipulationStarting(_this, outerArgs);
                }
                if (_this._viewerManipulationStarting) {
                    _this._viewerManipulationStarting(_this, outerArgs);
                }
            };
            this.i.viewerManipulationStarting = delegateCombine(this.i.viewerManipulationStarting, this._viewerManipulationStarting_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "viewerManipulationEnding", {
        /**
         * Raised when an ongoing manipulation of the series viewer is ended.
        */
        get: function () {
            return this._viewerManipulationEnding;
        },
        set: function (ev) {
            var _this = this;
            if (this._viewerManipulationEnding_wrapped !== null) {
                this.i.viewerManipulationEnding = delegateRemove(this.i.viewerManipulationEnding, this._viewerManipulationEnding_wrapped);
                this._viewerManipulationEnding_wrapped = null;
                this._viewerManipulationEnding = null;
            }
            this._viewerManipulationEnding = ev;
            this._viewerManipulationEnding_wrapped = function (o, e) {
                var outerArgs = new IgrSeriesViewerManipulationEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeViewerManipulationEnding) {
                    _this.beforeViewerManipulationEnding(_this, outerArgs);
                }
                if (_this._viewerManipulationEnding) {
                    _this._viewerManipulationEnding(_this, outerArgs);
                }
            };
            this.i.viewerManipulationEnding = delegateCombine(this.i.viewerManipulationEnding, this._viewerManipulationEnding_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "windowRectChanged", {
        /**
         * Occurs just after the current SeriesViewer's window rectangle is changed.
        *
        * &lt;-- position: content member-->
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        * <IgrDataChart
        * 	width="100%"
        * 	height="100%"
        * 	dataSource={this.data}
        * 	windowRectChanged={this.onWindowRectChanged} >
        * <IgrCategoryXAxis name="xAxis" label="Year"/>
        * <IgrNumericYAxis name="yAxis" minimumValue={0} />
        * <IgrLineSeries name="series3" title="Russia"
        *     valueMemberPath="Russia"
        *     xAxisName="xAxis"
        *     yAxisName="yAxis"
        *     markerType={this.state.markersType} />
        * </IgrDataChart>
        * ```
        *
        * ```ts
        * public onWindowRectChanged =( s:any, e: RectChangedEventArgs) => {
        *
        *        }
        * ```
        */
        get: function () {
            return this._windowRectChanged;
        },
        set: function (ev) {
            var _this = this;
            if (this._windowRectChanged_wrapped !== null) {
                this.i.windowRectChanged = delegateRemove(this.i.windowRectChanged, this._windowRectChanged_wrapped);
                this._windowRectChanged_wrapped = null;
                this._windowRectChanged = null;
            }
            this._windowRectChanged = ev;
            this._windowRectChanged_wrapped = function (o, e) {
                var outerArgs = new IgrRectChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeWindowRectChanged) {
                    _this.beforeWindowRectChanged(_this, outerArgs);
                }
                if (_this._windowRectChanged) {
                    _this._windowRectChanged(_this, outerArgs);
                }
            };
            this.i.windowRectChanged = delegateCombine(this.i.windowRectChanged, this._windowRectChanged_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "sizeChanged", {
        /**
         * Occurs after the size of the series viewer changes.
        *
        * &lt;-- position: content member-->
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        *  <IgrDataChart
        *     dataSource={this.state.dataSource}
        * 	sizeChanged={this.onSizeChanged=} >
        *
        *     <IgrNumericXAxis name="xAxis" />
        *     <IgrNumericYAxis name="yAxis" />
        *
        *     <IgrLineSeries
        *      name="series1"
        *     xAxisName="xAxis"
        * 	yAxisName="yAxis"
        * 	valueMemberPath="Value" />
        *  </IgrDataChart>
        * ```
        *
        * ```ts
        * public onSizeChanged =(event: any, args: RectChangedEventArgs}){
        *
        * }
        * ```
        */
        get: function () {
            return this._sizeChanged;
        },
        set: function (ev) {
            var _this = this;
            if (this._sizeChanged_wrapped !== null) {
                this.i.sizeChanged = delegateRemove(this.i.sizeChanged, this._sizeChanged_wrapped);
                this._sizeChanged_wrapped = null;
                this._sizeChanged = null;
            }
            this._sizeChanged = ev;
            this._sizeChanged_wrapped = function (o, e) {
                var outerArgs = new IgrRectChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeSizeChanged) {
                    _this.beforeSizeChanged(_this, outerArgs);
                }
                if (_this._sizeChanged) {
                    _this._sizeChanged(_this, outerArgs);
                }
            };
            this.i.sizeChanged = delegateCombine(this.i.sizeChanged, this._sizeChanged_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "actualWindowRectChanged", {
        /**
         * Raised when the actual window rectangle of the SeriesViewer changes.
        */
        get: function () {
            return this._actualWindowRectChanged;
        },
        set: function (ev) {
            var _this = this;
            if (this._actualWindowRectChanged_wrapped !== null) {
                this.i.actualWindowRectChanged = delegateRemove(this.i.actualWindowRectChanged, this._actualWindowRectChanged_wrapped);
                this._actualWindowRectChanged_wrapped = null;
                this._actualWindowRectChanged = null;
            }
            this._actualWindowRectChanged = ev;
            this._actualWindowRectChanged_wrapped = function (o, e) {
                var outerArgs = new IgrRectChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeActualWindowRectChanged) {
                    _this.beforeActualWindowRectChanged(_this, outerArgs);
                }
                if (_this._actualWindowRectChanged) {
                    _this._actualWindowRectChanged(_this, outerArgs);
                }
            };
            this.i.actualWindowRectChanged = delegateCombine(this.i.actualWindowRectChanged, this._actualWindowRectChanged_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "gridAreaRectChanged", {
        /**
         * Occurs just after the current SeriesViewer's grid area rectangle is changed.
         * The grid area may change as the result of the SeriesViewer being resized, or
         * of an axis being added or changing size, possibly in another SeriesViewer.
        */
        get: function () {
            return this._gridAreaRectChanged;
        },
        set: function (ev) {
            var _this = this;
            if (this._gridAreaRectChanged_wrapped !== null) {
                this.i.gridAreaRectChanged = delegateRemove(this.i.gridAreaRectChanged, this._gridAreaRectChanged_wrapped);
                this._gridAreaRectChanged_wrapped = null;
                this._gridAreaRectChanged = null;
            }
            this._gridAreaRectChanged = ev;
            this._gridAreaRectChanged_wrapped = function (o, e) {
                var outerArgs = new IgrRectChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeGridAreaRectChanged) {
                    _this.beforeGridAreaRectChanged(_this, outerArgs);
                }
                if (_this._gridAreaRectChanged) {
                    _this._gridAreaRectChanged(_this, outerArgs);
                }
            };
            this.i.gridAreaRectChanged = delegateCombine(this.i.gridAreaRectChanged, this._gridAreaRectChanged_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "selectedSeriesItemsChanged", {
        /**
         * Raised when the SelectedItems collection has changed.
        */
        get: function () {
            return this._selectedSeriesItemsChanged;
        },
        set: function (ev) {
            var _this = this;
            if (this._selectedSeriesItemsChanged_wrapped !== null) {
                this.i.selectedSeriesItemsChanged = delegateRemove(this.i.selectedSeriesItemsChanged, this._selectedSeriesItemsChanged_wrapped);
                this._selectedSeriesItemsChanged_wrapped = null;
                this._selectedSeriesItemsChanged = null;
            }
            this._selectedSeriesItemsChanged = ev;
            this._selectedSeriesItemsChanged_wrapped = function (o, e) {
                var outerArgs = new IgrSeriesViewerSelectedSeriesItemsChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeSelectedSeriesItemsChanged) {
                    _this.beforeSelectedSeriesItemsChanged(_this, outerArgs);
                }
                if (_this._selectedSeriesItemsChanged) {
                    _this._selectedSeriesItemsChanged(_this, outerArgs);
                }
            };
            this.i.selectedSeriesItemsChanged = delegateCombine(this.i.selectedSeriesItemsChanged, this._selectedSeriesItemsChanged_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "focusedSeriesItemsChanged", {
        /**
         * Raised when the SelectedItems collection has changed.
        */
        get: function () {
            return this._focusedSeriesItemsChanged;
        },
        set: function (ev) {
            var _this = this;
            if (this._focusedSeriesItemsChanged_wrapped !== null) {
                this.i.focusedSeriesItemsChanged = delegateRemove(this.i.focusedSeriesItemsChanged, this._focusedSeriesItemsChanged_wrapped);
                this._focusedSeriesItemsChanged_wrapped = null;
                this._focusedSeriesItemsChanged = null;
            }
            this._focusedSeriesItemsChanged = ev;
            this._focusedSeriesItemsChanged_wrapped = function (o, e) {
                var outerArgs = new IgrSeriesViewerSelectedSeriesItemsChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeFocusedSeriesItemsChanged) {
                    _this.beforeFocusedSeriesItemsChanged(_this, outerArgs);
                }
                if (_this._focusedSeriesItemsChanged) {
                    _this._focusedSeriesItemsChanged(_this, outerArgs);
                }
            };
            this.i.focusedSeriesItemsChanged = delegateCombine(this.i.focusedSeriesItemsChanged, this._focusedSeriesItemsChanged_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "refreshCompleted", {
        /**
         * Raised when the SeriesViewer's processing for an update has completed.
        */
        get: function () {
            return this._refreshCompleted;
        },
        set: function (ev) {
            var _this = this;
            if (this._refreshCompleted_wrapped !== null) {
                this.i.refreshCompleted = delegateRemove(this.i.refreshCompleted, this._refreshCompleted_wrapped);
                this._refreshCompleted_wrapped = null;
                this._refreshCompleted = null;
            }
            this._refreshCompleted = ev;
            this._refreshCompleted_wrapped = function (o, e) {
                var outerArgs = new IgrRefreshCompletedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeRefreshCompleted) {
                    _this.beforeRefreshCompleted(_this, outerArgs);
                }
                if (_this._refreshCompleted) {
                    _this._refreshCompleted(_this, outerArgs);
                }
            };
            this.i.refreshCompleted = delegateCombine(this.i.refreshCompleted, this._refreshCompleted_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSeriesViewer.prototype, "imageCaptured", {
        /**
         * Event raised when an image of the component is saved
        */
        get: function () {
            return this._imageCaptured;
        },
        set: function (ev) {
            var _this = this;
            if (this._imageCaptured_wrapped !== null) {
                this.i.imageCaptured = delegateRemove(this.i.imageCaptured, this._imageCaptured_wrapped);
                this._imageCaptured_wrapped = null;
                this._imageCaptured = null;
            }
            this._imageCaptured = ev;
            this._imageCaptured_wrapped = function (o, e) {
                var outerArgs = new IgrImageCapturedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeImageCaptured) {
                    _this.beforeImageCaptured(_this, outerArgs);
                }
                if (_this._imageCaptured) {
                    _this._imageCaptured(_this, outerArgs);
                }
            };
            this.i.imageCaptured = delegateCombine(this.i.imageCaptured, this._imageCaptured_wrapped);
            ;
        },
        enumerable: false,
        configurable: true
    });
    return IgrSeriesViewer;
}(React.Component));
export { IgrSeriesViewer };
