// igniteui-react-charts/esm5/lib/IndicatorProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, typeCast, markType, getInstanceType } from "igniteui-react-core";
import { IIndicatorProxy_$type } from "./IIndicatorProxy";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
/**
 * @hidden
 */
var IndicatorProxy = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IndicatorProxy, _super);
    function IndicatorProxy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    IndicatorProxy.prototype.matchesType = function (a) {
        if (a.e7) {
            return true;
        }
        return false;
    };
    IndicatorProxy.prototype.setCloseMemberPath = function (a, b) {
        a.za = b;
    };
    IndicatorProxy.prototype.setDisplayType = function (a, b) {
        a.aa3 = b;
    };
    IndicatorProxy.prototype.setHighMemberPath = function (a, b) {
        a.zn = b;
    };
    IndicatorProxy.prototype.setLongPeriod = function (a, b) {
        if (typeCast(IShortPeriodLongPeriod_$type, a) !== null) {
            a.longPeriod = b;
        }
    };
    IndicatorProxy.prototype.setLowMemberPath = function (a, b) {
        a.zs = b;
    };
    IndicatorProxy.prototype.setNegativeBrush = function (a, b) {
        a.aaw = b;
    };
    IndicatorProxy.prototype.setOpenMemberPath = function (a, b) {
        a.zw = b;
    };
    IndicatorProxy.prototype.setPeriodOverride = function (a, b) {
        if (typeCast(StrategyBasedIndicator.$, a) !== null) {
            a.aca(b);
        }
    };
    IndicatorProxy.prototype.setShortPeriod = function (a, b) {
        if (typeCast(IShortPeriodLongPeriod_$type, a) !== null) {
            a.shortPeriod = b;
        }
    };
    IndicatorProxy.prototype.setSignalPeriod = function (a, b) {
        if (getInstanceType(a).typeName == "MovingAverageConvergenceDivergenceIndicator") {
            var s_ = a;
            var p_ = b;
            s_['signalPeriod'] = p_;
        }
    };
    IndicatorProxy.prototype.setSmoothingPeriod = function (a, b) {
        if (getInstanceType(a).typeName == "FullStochasticOscillatorIndicator") {
            var s_ = a;
            var p_ = b;
            s_['smoothingPeriod'] = p_;
        }
    };
    IndicatorProxy.prototype.setVolumeMemberPath = function (a, b) {
        a.z0 = b;
    };
    IndicatorProxy.prototype.setXAxis = function (a, b) {
        a.w6 = b;
    };
    IndicatorProxy.prototype.setYAxis = function (a, b) {
        a.xp = b;
    };
    IndicatorProxy.prototype.handleCustomIndicatorEvent = function (a, b) {
        if (getInstanceType(a).typeName == "CustomIndicator") {
            var s_ = a;
            var p_ = b;
            s_['indicator'] = p_;
        }
    };
    IndicatorProxy.prototype.setMultiplier = function (a, b) {
        if (getInstanceType(a).typeName == "BollingerBandWidthIndicator") {
            var s_ = a;
            var p_ = b;
            s_['multiplier'] = p_;
        }
    };
    IndicatorProxy.$t = markType(IndicatorProxy, 'IndicatorProxy', Base.$, [IIndicatorProxy_$type]);
    return IndicatorProxy;
}(Base));
export { IndicatorProxy };
