// igniteui-react-charts/esm5/lib/LogarithmicScaler.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { NumericScaler } from "./NumericScaler";
import { markType } from "igniteui-react-core";
import { AutoRangeCalculator } from "./AutoRangeCalculator";
/**
 * @hidden
 */
var LogarithmicScaler = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(LogarithmicScaler, _super);
    function LogarithmicScaler() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.ae = 0;
        _this.ad = 0;
        return _this;
    }
    LogarithmicScaler.prototype.w = function (a, b, c) {
        _super.prototype.w.call(this, a, b, c);
        switch (a) {
            case NumericScaler.$$p[1]:
                this.ae = Math.log(this.l);
                break;
            case NumericScaler.$$p[0]:
                this.ad = Math.log(this.k);
                break;
        }
    };
    LogarithmicScaler.prototype.t = function (a, b, c, d, e) {
        var f;
        var g;
        var h = AutoRangeCalculator.a(a, b, c, true, a.mm, f, g);
        f = h.p5;
        g = h.p6;
        d = f;
        e = g;
        return {
            p3: d,
            p4: e
        };
    };
    LogarithmicScaler.$t = markType(LogarithmicScaler, 'LogarithmicScaler', NumericScaler.$);
    return LogarithmicScaler;
}(NumericScaler));
export { LogarithmicScaler };
