// igniteui-react-charts/esm5/lib/MarkerType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available types of marker than can be displayed by default by series with markers.
 */
export var MarkerType = /*@__PURE__*/ (function (MarkerType) {
    /**
     * Marker hasn't been set.
     */
    MarkerType[MarkerType["Unset"] = 0] = "Unset";
    /**
     * No marker shapes.
     */
    MarkerType[MarkerType["None"] = 1] = "None";
    /**
     * Automatic marker shape.
     */
    MarkerType[MarkerType["Automatic"] = 2] = "Automatic";
    /**
     * Circle marker shape.
     */
    MarkerType[MarkerType["Circle"] = 3] = "Circle";
    /**
     * Flat-top triangle marker shape.
     */
    MarkerType[MarkerType["Triangle"] = 4] = "Triangle";
    /**
     * Flat-base triangle marker shape.
     */
    MarkerType[MarkerType["Pyramid"] = 5] = "Pyramid";
    /**
     * Square marker shape.
     */
    MarkerType[MarkerType["Square"] = 6] = "Square";
    /**
     * Diamond marker shape.
     */
    MarkerType[MarkerType["Diamond"] = 7] = "Diamond";
    /**
     * Pentagon marker shape.
     */
    MarkerType[MarkerType["Pentagon"] = 8] = "Pentagon";
    /**
     * Hexagon marker shape.
     */
    MarkerType[MarkerType["Hexagon"] = 9] = "Hexagon";
    /**
     * Four-pointed star marker shape.
     */
    MarkerType[MarkerType["Tetragram"] = 10] = "Tetragram";
    /**
     * Five-pointed star marker shape.
     */
    MarkerType[MarkerType["Pentagram"] = 11] = "Pentagram";
    /**
     * Six-pointed star marker shape.
     */
    MarkerType[MarkerType["Hexagram"] = 12] = "Hexagram";
    /**
     * Renders transparent circular markers.
     */
    MarkerType[MarkerType["Hidden"] = 13] = "Hidden";
    return MarkerType;
})({});
/**
 * @hidden
 */
export var MarkerType_$type = /*@__PURE__*/ markEnum('MarkerType', 'Unset,0|None,1|Automatic,2|Circle,3|Triangle,4|Pyramid,5|Square,6|Diamond,7|Pentagon,8|Hexagon,9|Tetragram,10|Pentagram,11|Hexagram,12|Hidden,13');
