// igniteui-react-charts/esm5/lib/MoneyFlowIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { MoneyFlowIndexIndicatorStrategy } from "./MoneyFlowIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var MoneyFlowIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(MoneyFlowIndexIndicator, _super);
    function MoneyFlowIndexIndicator() {
        return _super.call(this) || this;
    }
    MoneyFlowIndexIndicator.prototype.bj = function () {
        return new MoneyFlowIndexIndicator();
    };
    MoneyFlowIndexIndicator.prototype.get_aa2 = function () {
        return 1;
    };
    MoneyFlowIndexIndicator.prototype.get_abv = function () {
        return new MoneyFlowIndexIndicatorStrategy();
    };
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    MoneyFlowIndexIndicator.prototype.get_ab8 = function () {
        return MoneyFlowIndexIndicator.$;
    };
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(MoneyFlowIndexIndicator.ach));
        },
        set: function (a) {
            this.h(MoneyFlowIndexIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    MoneyFlowIndexIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    MoneyFlowIndexIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    MoneyFlowIndexIndicator.$t = markType(MoneyFlowIndexIndicator, 'MoneyFlowIndexIndicator', StrategyBasedIndicator.$);
    MoneyFlowIndexIndicator.ach = StrategyBasedIndicator.acd(14, MoneyFlowIndexIndicator.$);
    return MoneyFlowIndexIndicator;
}(StrategyBasedIndicator));
export { MoneyFlowIndexIndicator };
