// igniteui-react-charts/esm5/lib/MovingAverageConvergenceDivergenceIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { MovingAverageConvergenceDivergenceIndicatorStrategy } from "./MovingAverageConvergenceDivergenceIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
/**
 * @hidden
 */
var MovingAverageConvergenceDivergenceIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(MovingAverageConvergenceDivergenceIndicator, _super);
    function MovingAverageConvergenceDivergenceIndicator() {
        var _this = _super.call(this) || this;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 10;
        _this.trendLinePeriod = 9;
        return _this;
    }
    MovingAverageConvergenceDivergenceIndicator.prototype.bj = function () {
        return new MovingAverageConvergenceDivergenceIndicator();
    };
    MovingAverageConvergenceDivergenceIndicator.prototype.get_aa2 = function () {
        return 1;
    };
    MovingAverageConvergenceDivergenceIndicator.prototype.get_abv = function () {
        return new MovingAverageConvergenceDivergenceIndicatorStrategy();
    };
    Object.defineProperty(MovingAverageConvergenceDivergenceIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    MovingAverageConvergenceDivergenceIndicator.prototype.get_ab8 = function () {
        return MovingAverageConvergenceDivergenceIndicator.$;
    };
    Object.defineProperty(MovingAverageConvergenceDivergenceIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(MovingAverageConvergenceDivergenceIndicator.prototype, "shortPeriod", {
        get: function () {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.acj));
        },
        set: function (a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.acj, a);
        },
        enumerable: false,
        configurable: true
    });
    MovingAverageConvergenceDivergenceIndicator.prototype.ab3 = function () {
        return this.shortPeriod;
    };
    Object.defineProperty(MovingAverageConvergenceDivergenceIndicator.prototype, "longPeriod", {
        get: function () {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.aci));
        },
        set: function (a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.aci, a);
        },
        enumerable: false,
        configurable: true
    });
    MovingAverageConvergenceDivergenceIndicator.prototype.ab0 = function () {
        return this.longPeriod;
    };
    Object.defineProperty(MovingAverageConvergenceDivergenceIndicator.prototype, "signalPeriod", {
        get: function () {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.ack));
        },
        set: function (a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.ack, a);
        },
        enumerable: false,
        configurable: true
    });
    MovingAverageConvergenceDivergenceIndicator.prototype.abb = function () {
        return this.signalPeriod;
    };
    MovingAverageConvergenceDivergenceIndicator.$t = markType(MovingAverageConvergenceDivergenceIndicator, 'MovingAverageConvergenceDivergenceIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    MovingAverageConvergenceDivergenceIndicator.acj = StrategyBasedIndicator.acf(10, MovingAverageConvergenceDivergenceIndicator.$);
    MovingAverageConvergenceDivergenceIndicator.aci = StrategyBasedIndicator.acb(30, MovingAverageConvergenceDivergenceIndicator.$);
    MovingAverageConvergenceDivergenceIndicator.ack = StrategyBasedIndicator.ace(9, MovingAverageConvergenceDivergenceIndicator.$, "SignalPeriod");
    return MovingAverageConvergenceDivergenceIndicator;
}(StrategyBasedIndicator));
export { MovingAverageConvergenceDivergenceIndicator };
