// igniteui-react-charts/esm5/lib/NormalRangeMessage_Spark.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { RenderingMessage_Spark } from "./RenderingMessage_Spark";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
var NormalRangeMessage_Spark = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NormalRangeMessage_Spark, _super);
    function NormalRangeMessage_Spark() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.g = 0;
        _this.h = 0;
        _this.f = 0;
        _this.e = 0;
        _this.c = false;
        return _this;
    }
    Object.defineProperty(NormalRangeMessage_Spark.prototype, "k", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NormalRangeMessage_Spark.prototype, "l", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NormalRangeMessage_Spark.prototype, "j", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NormalRangeMessage_Spark.prototype, "i", {
        get: function () {
            return this.e;
        },
        set: function (a) {
            this.e = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NormalRangeMessage_Spark.prototype, "d", {
        get: function () {
            return this.c;
        },
        set: function (a) {
            this.c = a;
        },
        enumerable: false,
        configurable: true
    });
    NormalRangeMessage_Spark.$t = markType(NormalRangeMessage_Spark, 'NormalRangeMessage_Spark', RenderingMessage_Spark.$);
    return NormalRangeMessage_Spark;
}(RenderingMessage_Spark));
export { NormalRangeMessage_Spark };
