// igniteui-react-charts/esm5/lib/NumericXAxis_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StraightNumericAxisBase } from "./StraightNumericAxisBase";
import { IScaler_$type } from "./IScaler";
import { markType, TypeRegistrar } from "igniteui-react-core";
import { HorizontalAxisLabelPanel } from "./HorizontalAxisLabelPanel";
import { ScalerParams } from "./ScalerParams";
import { Rect } from "igniteui-react-core";
import { RangeInfo } from "./RangeInfo";
import { HorizontalLogarithmicScaler } from "./HorizontalLogarithmicScaler";
import { FastHorizontalLinearScaler } from "./FastHorizontalLinearScaler";
import { isNaN_ } from "igniteui-react-core";
import { StraightNumericAxisBaseView } from "./StraightNumericAxisBaseView";
import { AxisDefaults } from "./AxisDefaults";
/**
 * @hidden
 */
var NumericXAxis = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NumericXAxis, _super);
    function NumericXAxis() {
        var _this = _super.call(this) || this;
        _this.ab = NumericXAxis.$;
        _this.k4 = _this.k2();
        return _this;
    }
    NumericXAxis.prototype.a4 = function () {
        return new NumericXAxisView(this);
    };
    NumericXAxis.prototype.hv = function (a) {
        _super.prototype.hv.call(this, a);
        this.oj = a;
    };
    NumericXAxis.prototype.av = function () {
        if (this.c6) {
            return (TypeRegistrar.create('HorizontalSmartAxisLabelPanel'));
        }
        else {
            return new HorizontalAxisLabelPanel();
        }
    };
    NumericXAxis.prototype.k2 = function () {
        var _this = this;
        var a = _super.prototype.k2.call(this);
        a.b.c = function (b) {
            if ((_this.az == null || _this.az.visibility == 0) && _this.aj != null) {
                _this.aw.bg = b;
                if (_this.az != null && (_this.az.location == 4 || _this.az.location == 5)) {
                    _this.bi.n3();
                }
            }
        };
        a.k = function (b, c, d) { return _this.iq(c, d, b.aa, b.b, false); };
        a.l = function (b, c, d, e) { return _this.ir(c, d, e, b.aa); };
        a.m = function (b, c) {
            var d = b.z;
            var e = new ScalerParams(0, b.ab, b.aa, _this.ci, d);
            return _this.el(c, e);
        };
        a.q = function (b, c) { return true; };
        a.j = function (b) { return _this.ht(b.u, b.n, b.aa, b.a, true); };
        a.e = function (b) {
            b.n = b.aa.bottom;
            if (_this.aj != null && _this.aj.bi != null) {
                var c = b.n;
                var d = b.r;
                var e = _this.ii(c, d, b.ab, b.aa, b.z, b.aa.top, b.aa.bottom);
                c = e.p0;
                d = e.p1;
                b.n = c;
                b.r = d;
            }
        };
        a.p = function (b, c, d) {
            var e = Math.round(c);
            return e >= Math.floor(b.aa.left) && e <= Math.ceil(b.aa.right);
        };
        return a;
    };
    NumericXAxis.prototype.el = function (a, b) {
        if (this.n2 != null) {
            return this.n2.p(a, b);
        }
        return this.n1.p(a, b);
    };
    NumericXAxis.prototype.eo = function (a, b) {
        if (this.n2 != null) {
            return this.n2.q(a, b);
        }
        return this.n1.q(a, b);
    };
    NumericXAxis.prototype.hq = function (a, b, c, d) {
        if (this.n2 != null) {
            this.n2.v(a, b, c, d);
            return;
        }
        this.n1.v(a, b, c, d);
    };
    NumericXAxis.prototype.hp = function (a, b, c, d, e) {
        if (this.n2 != null) {
            this.n2.u(a, b, c, d, e);
            return;
        }
        this.n1.u(a, b, c, d, e);
    };
    NumericXAxis.prototype.hr = function (a, b, c, d) {
        if (this.n2 != null) {
            this.n2.j(a, b, c, d);
            return;
        }
        this.n1.j(a, b, c, d);
    };
    NumericXAxis.prototype.ok = function (a, b) {
        if (this.bi == null || isNaN_(a) || isNaN_(b)) {
            return;
        }
        var c = this.n7(a, this);
        var d = this.n7(b, this);
        if (isNaN_(c) || isNaN_(d)) {
            return;
        }
        this.bi.wl = new Rect(0, c, this.bi.wl.y, d - c, this.bi.wl.height);
    };
    NumericXAxis.prototype.k5 = function (a, b, c, d) {
        var e = d;
        var f = _super.prototype.k5.call(this, a, b, c, d);
        var g = new ScalerParams(0, b, a, this.ch, c);
        var h = NaN;
        var i = NaN;
        if (!this.ch && b.right == 1) {
            i = this.l3;
        }
        else if (this.ch && b.left == 0) {
            h = this.l3;
        }
        if (isNaN_(h)) {
            h = this.eo(e.left, g);
        }
        if (isNaN_(i)) {
            i = this.eo(e.right, g);
        }
        var j = Math.min(h, i);
        var k = Math.max(h, i);
        this.l7 = j;
        this.l6 = k;
        f.k.add(((function () {
            var $ret = new RangeInfo();
            $ret.d = j;
            $ret.c = k;
            $ret.b = e.width;
            return $ret;
        })()));
        return f;
    };
    NumericXAxis.prototype.ic = function (a) {
        _super.prototype.ic.call(this, a);
        var b = this.bi != null ? this.bi.v4 : Rect.empty;
        var c = !b.isEmpty ? this.kh : Rect.empty;
        var d = this.kc();
        var e = this.kb();
        var f = this.k4.c(a, c, b, d, e);
        if (f != null) {
            this.l2 = f.f.m;
            this.l5 = f.f.l;
        }
    };
    NumericXAxis.prototype.is = function (a, b) {
        _super.prototype.is.call(this, a, b);
        if (b.height != a.height) {
            this.c1();
        }
    };
    NumericXAxis.prototype.get_a2 = function () {
        return 0;
    };
    Object.defineProperty(NumericXAxis.prototype, "a2", {
        get: function () {
            return this.get_a2();
        },
        enumerable: false,
        configurable: true
    });
    NumericXAxis.prototype.k7 = function () {
        if (this.lo) {
            return new HorizontalLogarithmicScaler();
        }
        switch (this.n0) {
            case 0: return new FastHorizontalLinearScaler();
            case 1: return new HorizontalLogarithmicScaler();
        }
        return null;
    };
    NumericXAxis.prototype.j$d = function (a, b) {
        return this.eo.apply(this, arguments);
    };
    NumericXAxis.prototype.j$c = function (a, b) {
        return this.el.apply(this, arguments);
    };
    NumericXAxis.prototype.j$e = function (a, b, c, d, e) {
        this.hp.apply(this, arguments);
    };
    NumericXAxis.prototype.j$f = function (a, b, c, d) {
        this.hq.apply(this, arguments);
    };
    NumericXAxis.prototype.j$g = function (a, b, c, d) {
        this.hr.apply(this, arguments);
    };
    Object.defineProperty(NumericXAxis.prototype, "j$b", {
        get: function () {
            return this.ch;
        },
        enumerable: false,
        configurable: true
    });
    NumericXAxis.prototype.j$a = function () {
        return this.ag.apply(this, arguments);
    };
    NumericXAxis.$t = markType(NumericXAxis, 'NumericXAxis', StraightNumericAxisBase.$, [IScaler_$type]);
    return NumericXAxis;
}(StraightNumericAxisBase));
export { NumericXAxis };
/**
 * @hidden
 */
var NumericXAxisView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NumericXAxisView, _super);
    function NumericXAxisView(a) {
        var _this = _super.call(this, a) || this;
        _this.bo = null;
        _this.bo = a;
        return _this;
    }
    NumericXAxisView.prototype.al = function () {
        _super.prototype.al.call(this);
        this.a.jz = AxisDefaults.h;
    };
    NumericXAxisView.$t = markType(NumericXAxisView, 'NumericXAxisView', StraightNumericAxisBaseView.$);
    return NumericXAxisView;
}(StraightNumericAxisBaseView));
export { NumericXAxisView };
