// igniteui-react-charts/esm5/lib/OverlayProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, markType, getInstanceType } from "igniteui-react-core";
import { IOverlayProxy_$type } from "./IOverlayProxy";
/**
 * @hidden
 */
var OverlayProxy = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(OverlayProxy, _super);
    function OverlayProxy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OverlayProxy.prototype.matchesType = function (a) {
        if (a.e8) {
            return true;
        }
        return false;
    };
    OverlayProxy.prototype.setCloseMemberPath = function (a, b) {
        a.za = b;
    };
    OverlayProxy.prototype.setHighMemberPath = function (a, b) {
        a.zn = b;
    };
    OverlayProxy.prototype.setLowMemberPath = function (a, b) {
        a.zs = b;
    };
    OverlayProxy.prototype.setMultiplier = function (a, b) {
        if (getInstanceType(a).typeName == "BollingerBandsOverlay") {
            var s_ = a;
            var p_ = b;
            s_['multiplier'] = p_;
        }
    };
    OverlayProxy.prototype.setOpenMemberPath = function (a, b) {
        a.zw = b;
    };
    OverlayProxy.prototype.setVolumeMemberPath = function (a, b) {
        a.z0 = b;
    };
    OverlayProxy.prototype.setXAxis = function (a, b) {
        a.w6 = b;
    };
    OverlayProxy.prototype.setYAxis = function (a, b) {
        a.xp = b;
    };
    OverlayProxy.$t = markType(OverlayProxy, 'OverlayProxy', Base.$, [IOverlayProxy_$type]);
    return OverlayProxy;
}(Base));
export { OverlayProxy };
