// igniteui-react-charts/esm5/lib/PercentageVolumeOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentageVolumeOscillatorIndicatorStrategy } from "./PercentageVolumeOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var PercentageVolumeOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PercentageVolumeOscillatorIndicator, _super);
    function PercentageVolumeOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PercentageVolumeOscillatorIndicator.prototype.bj = function () {
        return new PercentageVolumeOscillatorIndicator();
    };
    PercentageVolumeOscillatorIndicator.prototype.get_abv = function () {
        return new PercentageVolumeOscillatorIndicatorStrategy();
    };
    Object.defineProperty(PercentageVolumeOscillatorIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    PercentageVolumeOscillatorIndicator.prototype.get_ab8 = function () {
        return PercentageVolumeOscillatorIndicator.$;
    };
    Object.defineProperty(PercentageVolumeOscillatorIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PercentageVolumeOscillatorIndicator.prototype, "shortPeriod", {
        get: function () {
            return typeGetValue(this.c(PercentageVolumeOscillatorIndicator.ach));
        },
        set: function (a) {
            this.h(PercentageVolumeOscillatorIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    PercentageVolumeOscillatorIndicator.prototype.ab3 = function () {
        return this.shortPeriod;
    };
    Object.defineProperty(PercentageVolumeOscillatorIndicator.prototype, "longPeriod", {
        get: function () {
            return typeGetValue(this.c(PercentageVolumeOscillatorIndicator.acg));
        },
        set: function (a) {
            this.h(PercentageVolumeOscillatorIndicator.acg, a);
        },
        enumerable: false,
        configurable: true
    });
    PercentageVolumeOscillatorIndicator.prototype.ab0 = function () {
        return this.longPeriod;
    };
    PercentageVolumeOscillatorIndicator.$t = markType(PercentageVolumeOscillatorIndicator, 'PercentageVolumeOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    PercentageVolumeOscillatorIndicator.ach = StrategyBasedIndicator.acf(10, PercentageVolumeOscillatorIndicator.$);
    PercentageVolumeOscillatorIndicator.acg = StrategyBasedIndicator.acb(30, PercentageVolumeOscillatorIndicator.$);
    return PercentageVolumeOscillatorIndicator;
}(StrategyBasedIndicator));
export { PercentageVolumeOscillatorIndicator };
