// igniteui-react-charts/esm5/lib/PercentagePriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentagePriceOscillatorIndicatorStrategy } from "./PercentagePriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var PercentagePriceOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PercentagePriceOscillatorIndicator, _super);
    function PercentagePriceOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PercentagePriceOscillatorIndicator.prototype.bj = function () {
        return new PercentagePriceOscillatorIndicator();
    };
    PercentagePriceOscillatorIndicator.prototype.get_abv = function () {
        return new PercentagePriceOscillatorIndicatorStrategy();
    };
    Object.defineProperty(PercentagePriceOscillatorIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    PercentagePriceOscillatorIndicator.prototype.get_ab8 = function () {
        return PercentagePriceOscillatorIndicator.$;
    };
    Object.defineProperty(PercentagePriceOscillatorIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PercentagePriceOscillatorIndicator.prototype, "shortPeriod", {
        get: function () {
            return typeGetValue(this.c(PercentagePriceOscillatorIndicator.ach));
        },
        set: function (a) {
            this.h(PercentagePriceOscillatorIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    PercentagePriceOscillatorIndicator.prototype.ab3 = function () {
        return this.shortPeriod;
    };
    Object.defineProperty(PercentagePriceOscillatorIndicator.prototype, "longPeriod", {
        get: function () {
            return typeGetValue(this.c(PercentagePriceOscillatorIndicator.acg));
        },
        set: function (a) {
            this.h(PercentagePriceOscillatorIndicator.acg, a);
        },
        enumerable: false,
        configurable: true
    });
    PercentagePriceOscillatorIndicator.prototype.ab0 = function () {
        return this.longPeriod;
    };
    PercentagePriceOscillatorIndicator.$t = markType(PercentagePriceOscillatorIndicator, 'PercentagePriceOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    PercentagePriceOscillatorIndicator.ach = StrategyBasedIndicator.acf(10, PercentagePriceOscillatorIndicator.$);
    PercentagePriceOscillatorIndicator.acg = StrategyBasedIndicator.acb(30, PercentagePriceOscillatorIndicator.$);
    return PercentagePriceOscillatorIndicator;
}(StrategyBasedIndicator));
export { PercentagePriceOscillatorIndicator };
