// igniteui-react-charts/esm5/lib/PointSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { HorizontalAnchoredCategorySeries } from "./HorizontalAnchoredCategorySeries";
import { markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { ChartSelection } from "./ChartSelection";
import { AnchoredCategorySeriesView } from "./AnchoredCategorySeriesView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
var PointSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PointSeries, _super);
    function PointSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PointSeries.$;
        return _this;
    }
    PointSeries.prototype.bj = function () {
        return new PointSeries();
    };
    PointSeries.prototype.get_ei = function () {
        return true;
    };
    PointSeries.prototype.cs = function () {
        return new PointSeriesView(this);
    };
    PointSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.acw = a;
    };
    PointSeries.prototype.jl = function () {
        return 0.7;
    };
    PointSeries.prototype.w1 = function () {
        if (this.w0 == 0) {
            return 2;
        }
        return _super.prototype.w1.call(this);
    };
    PointSeries.prototype.gw = function (a, b) {
        var c = this.wc(this.wh(a), true, false);
        if (this.xh(a, b)) {
            return true;
        }
        return false;
    };
    PointSeries.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
    };
    PointSeries.prototype.p4 = function (a, b, c) {
        this.aac(a, b, c, this.yt, 0);
    };
    PointSeries.prototype.ak = function (a, b) {
        var c = new List$1(ChartSelection.$, 0);
        this.p4(a, c, b);
        return c.toArray();
    };
    Object.defineProperty(PointSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zq;
        },
        enumerable: false,
        configurable: true
    });
    PointSeries.$t = markType(PointSeries, 'PointSeries', HorizontalAnchoredCategorySeries.$);
    return PointSeries;
}(HorizontalAnchoredCategorySeries));
export { PointSeries };
/**
 * @hidden
 */
var PointSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PointSeriesView, _super);
    function PointSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.df = null;
        _this.df = a;
        return _this;
    }
    PointSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.b4.xa = Defaults.e;
        }
    };
    PointSeriesView.prototype.af = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.v1 = a;
        this.e.jm = 3;
        this.e.jn = 2;
        this.e.jo = 2;
        this.e.hc = false;
    };
    PointSeriesView.$t = markType(PointSeriesView, 'PointSeriesView', AnchoredCategorySeriesView.$);
    return PointSeriesView;
}(AnchoredCategorySeriesView));
export { PointSeriesView };
