// igniteui-react-charts/esm5/lib/PolarScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { PolarBase } from "./PolarBase";
import { markType } from "igniteui-react-core";
import { PolarBaseView } from "./PolarBaseView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
var PolarScatterSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarScatterSeries, _super);
    function PolarScatterSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PolarScatterSeries.$;
        return _this;
    }
    PolarScatterSeries.prototype.bj = function () {
        return new PolarScatterSeries();
    };
    PolarScatterSeries.prototype.get_ei = function () {
        return true;
    };
    PolarScatterSeries.prototype.cs = function () {
        return new PolarScatterSeriesView(this);
    };
    PolarScatterSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.acg = a;
    };
    PolarScatterSeries.prototype.jl = function () {
        return 0.7;
    };
    PolarScatterSeries.prototype.w1 = function () {
        if (this.w0 == 0) {
            return 2;
        }
        return _super.prototype.w1.call(this);
    };
    PolarScatterSeries.$t = markType(PolarScatterSeries, 'PolarScatterSeries', PolarBase.$);
    return PolarScatterSeries;
}(PolarBase));
export { PolarScatterSeries };
/**
 * @hidden
 */
var PolarScatterSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarScatterSeriesView, _super);
    function PolarScatterSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c8 = null;
        _this.c8 = a;
        return _this;
    }
    PolarScatterSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c8.xa = Defaults.e;
        }
    };
    PolarScatterSeriesView.prototype.af = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.v1 = a;
        this.e.jm = 3;
        this.e.jn = 2;
        this.e.jo = 2;
        this.e.hc = false;
    };
    PolarScatterSeriesView.$t = markType(PolarScatterSeriesView, 'PolarScatterSeriesView', PolarBaseView.$);
    return PolarScatterSeriesView;
}(PolarBaseView));
export { PolarScatterSeriesView };
