// igniteui-react-charts/esm5/lib/PolarSplineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { DependencyProperty } from "igniteui-react-core";
import { runOn, markType, markDep } from "igniteui-react-core";
import { Path } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { PathGeometry } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { PathFigure } from "igniteui-react-core";
import { Extensions } from "igniteui-react-core";
import { Numeric } from "igniteui-react-core";
import { Color } from "igniteui-react-core";
/**
 * @hidden
 */
var PolarSplineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarSplineSeries, _super);
    function PolarSplineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PolarSplineSeries.$;
        return _this;
    }
    PolarSplineSeries.prototype.bj = function () {
        return new PolarSplineSeries();
    };
    PolarSplineSeries.prototype.get_f0 = function () {
        return true;
    };
    PolarSplineSeries.prototype.cs = function () {
        return new PolarSplineSeriesView(this);
    };
    PolarSplineSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.acu = a;
    };
    PolarSplineSeries.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case PolarSplineSeries.$$p[0]:
                this.rr(false);
                break;
        }
    };
    PolarSplineSeries.prototype.acs = function (a) {
        var b = a;
        b.da();
    };
    PolarSplineSeries.prototype.act = function (a, b) {
        this.acs(b);
        var c = b;
        c.db(a.m, this.acv);
        this.y5.az(this, this.zx(), runOn(this, this.yn));
        var d = false;
        var e = this.y5.b;
        if (e != null) {
            d = true;
        }
        if (d) {
            this.abc(-1, this.zj.count, b.q);
        }
        this.y5.ad(b.dd, true, false, true, false);
    };
    PolarSplineSeries.$t = markType(PolarSplineSeries, 'PolarSplineSeries', PolarLineSeriesBase.$);
    PolarSplineSeries.$$p = markDep(DependencyProperty, PropertyMetadata, PolarSplineSeries, 'raisePropertyChanged', ['Stiffness:acv:acx', [1, 0.5]]);
    return PolarSplineSeries;
}(PolarLineSeriesBase));
export { PolarSplineSeries };
/**
 * @hidden
 */
var PolarSplineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarSplineSeriesView, _super);
    function PolarSplineSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c9 = null;
        _this.dd = ((function () {
            var $ret = new Path();
            $ret.an = new PathGeometry();
            return $ret;
        })());
        _this.dc = ((function () {
            var $ret = new Path();
            $ret.an = new PathGeometry();
            return $ret;
        })());
        _this.c9 = a;
        return _this;
    }
    PolarSplineSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.b4.xa = Defaults.e;
        }
        var a = new PathGeometry();
        a.c.add(((function () {
            var $ret = new PathFigure();
            return $ret;
        })()));
        this.dd.an = a;
    };
    PolarSplineSeriesView.prototype.da = function () {
        Extensions.e(this.dd.an);
        Extensions.e(this.dc.an);
        this.as();
    };
    PolarSplineSeriesView.prototype.db = function (a, b) {
        this.dd.an.c = Numeric.h(a.count, function (c) { return a._inner[c].x; }, function (c) { return a._inner[c].y; }, b);
        this.as();
    };
    PolarSplineSeriesView.prototype.bd = function () {
        _super.prototype.bd.call(this);
    };
    PolarSplineSeriesView.prototype.be = function () {
        _super.prototype.be.call(this);
        var a = this.br();
        this.dc.an = this.dd.an;
        this.dc._stroke = a;
        this.dc.ad = this.e.ie + 3;
    };
    PolarSplineSeriesView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (a.d) {
            if (b) {
                a.t(this.dc);
            }
            else {
                a.t(this.dd);
            }
        }
    };
    PolarSplineSeriesView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        this.e.bc.exportPathData(a, this.dd, "polarShape", ["Main"]);
    };
    PolarSplineSeriesView.prototype.af = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.v1 = a;
        this.e.jm = 3;
        this.e.jn = 1;
        this.e.jo = 4;
        this.e.hc = false;
    };
    PolarSplineSeriesView.$t = markType(PolarSplineSeriesView, 'PolarSplineSeriesView', PolarLineSeriesBaseView.$);
    return PolarSplineSeriesView;
}(PolarLineSeriesBaseView));
export { PolarSplineSeriesView };
