// igniteui-react-charts/esm5/lib/PolarBase_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { MarkerSeries } from "./MarkerSeries";
import { Base, runOn, delegateCombine, Number_$type, enumGetBox, Point_$type, typeCast, fromEnum, markType, markDep } from "igniteui-react-core";
import { PolarAxisInfoCache } from "./PolarAxisInfoCache";
import { SeriesRenderer$2 } from "./SeriesRenderer$2";
import { PolarFrame } from "./PolarFrame";
import { DependencyProperty } from "igniteui-react-core";
import { NumericAngleAxis } from "./NumericAngleAxis";
import { NumericRadiusAxis } from "./NumericRadiusAxis";
import { TrendLineType_$type } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
import { PenLineCap_$type } from "igniteui-react-core";
import { DoubleCollection } from "igniteui-react-core";
import { ScatterItemSearchMode_$type } from "./ScatterItemSearchMode";
import { HighlightingInfo } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { ScalerParams } from "./ScalerParams";
import { Rect } from "igniteui-react-core";
import { Marker } from "./Marker";
import { MarkerSeriesView } from "./MarkerSeriesView";
import { PolarAxes } from "./PolarAxes";
import { SeriesLayerPropertyOverlay } from "./SeriesLayerPropertyOverlay";
import { IFastItemsSource_$type } from "igniteui-react-core";
import { SeriesRenderingArguments } from "./SeriesRenderingArguments";
import { AxisRange } from "./AxisRange";
import { Dictionary$2 } from "igniteui-react-core";
import { OwnedPoint } from "./OwnedPoint";
import { GeometryUtil } from "igniteui-react-core";
import { DataContext } from "igniteui-react-core";
import { HashPool$2 } from "igniteui-react-core";
import { Clipper } from "igniteui-react-core";
import { PolarTrendLineManager } from "./PolarTrendLineManager";
import { TrendResolutionParams } from "./TrendResolutionParams";
import { AssigningPolarStyleEventArgs } from "./AssigningPolarStyleEventArgs";
import { AssigningPolarMarkerStyleEventArgs } from "./AssigningPolarMarkerStyleEventArgs";
import { SeriesRenderManager } from "./SeriesRenderManager";
import { DataLegendSeriesValueInfo } from "igniteui-react-core";
import { DataLegendSeriesContext } from "igniteui-react-core";
import { ChartSelection } from "./ChartSelection";
import { CollisionRect } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { truncate, isNaN_, isInfinity } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { NumericMarkerManager } from "./NumericMarkerManager";
import { GeometryGroup } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
/**
 * @hidden
 */
var PolarBase = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarBase, _super);
    function PolarBase() {
        var _this = _super.call(this) || this;
        _this.yy = null;
        _this.yp = null;
        _this.yw = null;
        _this.acf = null;
        _this.ace = null;
        _this.acd = null;
        _this.za = null;
        _this.zh = null;
        _this.zi = null;
        _this.y2 = new PolarFrame();
        _this.y4 = new PolarFrame();
        _this.y0 = new PolarFrame();
        _this.yv = null;
        _this.z0 = null;
        _this.y6 = 1;
        _this.zm = false;
        _this.zl = false;
        _this.zf = null;
        _this.zg = null;
        _this.assigningPolarStyle = null;
        _this.assigningPolarMarkerStyle = null;
        _this.y5 = null;
        _this.y3 = null;
        _this.yz = null;
        _this.y5 = new PolarSeriesRenderManager();
        _this.y7 = _this.y8();
        _this.y3 = new PolarFrame();
        _this.y4.x = runOn(_this, _this.acc);
        _this.y2.x = runOn(_this, _this.acc);
        _this.y0.x = runOn(_this, _this.acc);
        _this.za = new SeriesRenderer$2(PolarFrame.$, PolarBaseView.$, 0, runOn(_this, _this.abd), runOn(_this, _this.abh), runOn(_this, _this.d3), runOn(_this, _this.r7), runOn(_this, _this.aa6));
        return _this;
    }
    PolarBase.prototype.cs = function () {
        return new PolarBaseView(this);
    };
    PolarBase.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.yx = a;
    };
    PolarBase.prototype.o2 = function () {
        _super.prototype.o2.call(this);
        this.yt = null;
        this.yu = null;
    };
    PolarBase.prototype.get_fp = function () {
        return true;
    };
    PolarBase.prototype.eb = function (a, b, c, d, e) {
        var _this = this;
        if (!this.hasMarkers) {
            return false;
        }
        var f = this.yt;
        var g = new ScalerParams(1, this.db.v4, this.ws, f.ch);
        g.c = this.getEffectiveViewport1(this.cv);
        var h = this.yu;
        var i = new ScalerParams(1, this.db.v4, this.ws, h.ch);
        i.c = this.getEffectiveViewport1(this.cv);
        a.clear();
        this.w6.cj(function (j) {
            if (j._visibility == 0) {
                var k = new SeriesPointOfInterest();
                var l = j.n;
                var m = j.o;
                if (l < _this.ws.left || l > _this.ws.right || m < _this.ws.top || m > _this.ws.bottom) {
                    return;
                }
                var n = void 0;
                var o = void 0;
                var p = _this.yv.f(l, m, _this.cv.by, _this.cv.bx, _this.cv.getEffectiveViewport(), n, o);
                n = p.p5;
                o = p.p6;
                k.d = n;
                k.e = o;
                k.b = j.content;
                k.g = _this.vg;
                k.c = j.ak;
                if (_this.g9) {
                    k.g = _this.vn(j.ak);
                }
                a.add(k);
            }
        });
        return true;
    };
    PolarBase.prototype.aa6 = function () {
        if (this.dc.g) {
            this.dc.t();
        }
    };
    PolarBase.prototype.aa8 = function () {
        this.yv.e(this.yy.w, this.zj, this.zk, this.acf, this.ace, this.acd, function (a, b) { return Math.cos(b); }, function (a, b) { return Math.sin(b); }, true);
    };
    PolarBase.prototype.gb = function () {
        return this.highlightedItemsSource != null || (this.aao != null && this.aam != null);
    };
    PolarBase.prototype.zy = function () {
        return false;
    };
    PolarBase.prototype.ht = function () {
        var a = _super.prototype.ht.call(this);
        {
            var b = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "AngleMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedAngleMemberPath";
                $ret.q = true;
                return $ret;
            })());
            b.valueResolving = delegateCombine(b.valueResolving, runOn(this, this.aa9));
            a.add(b);
        }
        {
            var c = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "RadiusMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedRadiusMemberPath";
                $ret.q = true;
                return $ret;
            })());
            c.valueResolving = delegateCombine(c.valueResolving, runOn(this, this.aba));
            a.add(c);
        }
        if (this.zy()) {
            var d = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "DashArray";
                $ret.p = true;
                $ret.x = ((function () {
                    var $ret = new DoubleCollection();
                    $ret.add(5);
                    $ret.add(5);
                    return $ret;
                })());
                $ret.q = true;
                return $ret;
            })());
            a.add(d);
        }
        if (!this.zy()) {
            var e = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Brush";
                $ret.p = true;
                $ret.y = "ActualBrush";
                $ret.q = true;
                return $ret;
            })());
            e.valueResolving = delegateCombine(e.valueResolving, runOn(this, this.r0));
            a.add(e);
        }
        if (!this.zy()) {
            var f = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Outline";
                $ret.p = true;
                $ret.y = "ActualOutline";
                $ret.q = true;
                return $ret;
            })());
            f.valueResolving = delegateCombine(f.valueResolving, runOn(this, this.r0));
            a.add(f);
        }
        return a;
    };
    PolarBase.prototype.aba = function (a, b) {
        if (this.aao != null) {
            b.value = this.aao;
            return;
        }
        b.value = this.aa1;
    };
    PolarBase.prototype.aa9 = function (a, b) {
        if (this.aam != null) {
            b.value = this.aam;
            return;
        }
        b.value = this.aaj;
    };
    Object.defineProperty(PolarBase.prototype, "zj", {
        get: function () {
            return this.zh;
        },
        set: function (a) {
            if (this.zh != a) {
                var b = this.zj;
                this.zh = a;
                this.raisePropertyChanged("AngleColumn", b, this.zj);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBase.prototype, "zk", {
        get: function () {
            return this.zi;
        },
        set: function (a) {
            if (this.zi != a) {
                var b = this.zk;
                this.zi = a;
                this.raisePropertyChanged("RadiusColumn", b, this.zk);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBase.prototype, "ab7", {
        get: function () {
            return this.c(PolarBase.abi);
        },
        set: function (a) {
            this.h(PolarBase.abi, a);
        },
        enumerable: false,
        configurable: true
    });
    PolarBase.prototype.y1 = function () {
        var a = this.y0;
        if (this.d3()) {
            if (this.dc.g) {
                this.dc.t();
            }
            a = this.y4;
        }
        return a;
    };
    PolarBase.prototype.qb = function () {
        _super.prototype.qb.call(this);
        if (this.yu != null) {
            this.yu.ib(false);
        }
        if (this.yt != null) {
            this.yt.ib(false);
        }
    };
    PolarBase.prototype.pb = function (a, b, c, d) {
        var e = false;
        this.yx.c0.g$e(a, b, c, d);
        if (this.yt != null && !this.yt.c1()) {
            e = true;
        }
        if (this.yu != null && !this.yu.c1()) {
            e = true;
        }
        if (e) {
            this.rr(true);
        }
    };
    PolarBase.prototype.sq = function (a, b) {
        this.rr(false);
    };
    PolarBase.prototype.sr = function (a, b) {
        this.rr(false);
    };
    PolarBase.prototype.he = function (a, b, c) {
        var d = true;
        if (!_super.prototype.he.call(this, a, b, c) || !c.o() || b.isEmpty || a.isEmpty || this.yt == null || this.yu == null || this.zj == null || this.zk == null || this.zj.count == 0 || this.zk.count == 0 || this.dn == null || this.dn.count != this.zj.count || this.dn.count != this.zk.count || this.yv == null || this.yt.bi == null || this.yu.bi == null || this.yt.l4 == this.yt.l3 || this.yu.l4 == this.yu.l3) {
            d = false;
        }
        return d;
    };
    PolarBase.prototype.rs = function (a) {
        _super.prototype.rs.call(this, a);
        var b;
        var c;
        var d = this.getViewInfo(c, b);
        c = d.p0;
        b = d.p1;
        if (!this.he(c, b, this.cv)) {
            this.o5(true, this.cv);
            return;
        }
        this.acf = b;
        this.ace = c;
        this.acd = this.cv.getEffectiveViewport();
        this.yw = new PolarAxisInfoCache(this.yt, this.yu, this.dn);
        var e = new SeriesRenderingArguments(this, c, b, a, this.gr);
        var f = this.za.g(e, this.y2, this.y0, this.y4, this.yx);
        this.y2 = f.p1;
        this.y0 = f.p2;
        this.y4 = f.p3;
    };
    PolarBase.prototype.acc = function (a) {
        var b = this.yv.h(a.x, a.y, this.acf, this.ace, this.acd, this.yw.e, this.yw.d, this.yw.g, this.yw.f, this.yw.i, this.yw.h);
        return b;
    };
    PolarBase.prototype.ay = function (a) {
        if (a != null && a == this.yt && this.zj != null) {
            return new AxisRange(this.zj.minimum, this.zj.maximum);
        }
        if (a != null && a == this.yu && this.zk != null) {
            return new AxisRange(this.zk.minimum, this.zk.maximum);
        }
        return null;
    };
    PolarBase.prototype.gd = function (a) {
        var b = this.cv.by;
        b = b.copy();
        var c = this.cv.bx;
        var d = !b.isEmpty && !c.isEmpty && this.dn != null ? this.dn.indexOf(a) : -1;
        if (this.yt == null || this.zj == null || this.yu == null || this.zk == null) {
            return false;
        }
        if (d < 0 || d > this.zj.count - 1 || d > this.zk.count - 1) {
            return false;
        }
        var e = this.yt.getScaledAngle(this.zj.item(d));
        var f = this.yu.od(this.zk.item(d));
        var g = 0.5 + (Math.cos(e) * f);
        var h = 0.5 + (Math.sin(e) * f);
        if (!isNaN_(g)) {
            if (g < b.left + 0.1 * b.width) {
                g = g + 0.4 * b.width;
                b.x = g - 0.5 * b.width;
            }
            if (g > b.right - 0.1 * b.width) {
                g = g - 0.4 * b.width;
                b.x = g - 0.5 * b.width;
            }
        }
        if (!isNaN_(h)) {
            if (h < b.top + 0.1 * b.height) {
                h = h + 0.4 * b.height;
                b.y = h - 0.5 * b.height;
            }
            if (h > b.bottom - 0.1 * b.height) {
                h = h - 0.4 * b.height;
                b.y = h - 0.5 * b.height;
            }
        }
        if (this.c3 != null) {
            this.c3.au(this.db, b, true);
        }
        return d >= 0;
    };
    PolarBase.prototype.kn = function (a) {
        var b = this.j3(a);
        if (this.dn != null && b >= 0 && b < this.dn.count) {
            return this.dn.item(b);
        }
        return null;
    };
    PolarBase.prototype.j3 = function (a) {
        var b = truncate(Math.round(this.ir(a)));
        return b;
    };
    PolarBase.prototype.ir = function (a) {
        switch (this.y7) {
            case 1: return this.z4(a);
            case 3: return this.z5(a);
            case 2: return this.z2(a);
            case 4: return this.z3(a);
        }
        return -1;
    };
    PolarBase.prototype.z3 = function (a) {
        return this.z6(a, false);
    };
    PolarBase.prototype.z2 = function (a) {
        var b = this.v7(a);
        var c = b.x;
        var d = b.y;
        if (this.dn == null || this.yt == null || this.yu == null || this.zj == null || this.zk == null) {
            return -1;
        }
        if (this.dn.count > this.z8) {
            return -1;
        }
        if (!this.zq()) {
            return -1;
        }
        var e = this.z0;
        var f = 1.7976931348623157E+308;
        var g = -1;
        for (var h = 0; h < this.zj.count; h++) {
            var i = e._inner[h].x;
            var j = e._inner[h].y;
            var k = (c - i) * (c - i) + (d - j) * (d - j);
            if (k < f) {
                f = k;
                g = h;
            }
        }
        return g;
    };
    PolarBase.prototype.zq = function () {
        var a = this.y0;
        if (this.d3()) {
            a = this.y4;
        }
        if (a == null || a.w.count == 0) {
            return false;
        }
        this.z0 = a.w;
        return true;
    };
    PolarBase.prototype.z6 = function (a, b) {
        if (!this.zt) {
            return -1;
        }
        var c = { $type: Point_$type, x: 0, y: 0 };
        var d = true;
        var e = this.y0;
        if (this.d3()) {
            e = this.y4;
        }
        if (e == null) {
            return -1;
        }
        if (e.m == null) {
            return -1;
        }
        if (this.dn == null || this.dn.count > this.z8) {
            return -1;
        }
        if (!this.zq()) {
            return -1;
        }
        var f = this.z0;
        var g = this.v7(a);
        var h = 1.7976931348623157E+308;
        var i = -1;
        for (var j = 0; j < f.count; j++) {
            var k = f._inner[j];
            if (j > 0) {
                if (isNaN_(k.x) || isNaN_(k.y) || isNaN_(c.x) || isNaN_(c.y)) {
                    c = k;
                    continue;
                }
                if (!b || e.j.containsKey(this.dn.item(j)) || e.j.containsKey(this.dn.item(j - 1))) {
                    var l = GeometryUtil.h(g, c, k);
                    if (l < h) {
                        h = l;
                        i = j;
                    }
                }
            }
            c = k;
        }
        if (i >= 0) {
            var m = f._inner[i];
            var n = f._inner[i - 1];
            if (!b || (e.j.containsKey(this.dn.item(i)) && e.j.containsKey(this.dn.item(i - 1)))) {
                var o = GeometryUtil.m(g, n, m);
                return (i - 1) + o;
            }
            else if (e.j.containsKey(this.dn.item(i))) {
                return i;
            }
            else {
                return i - 1;
            }
        }
        return -1;
    };
    PolarBase.prototype.z5 = function (a) {
        return this.z6(a, true);
    };
    PolarBase.prototype.z4 = function (a) {
        var _this = this;
        var b = null;
        var c = 1.7976931348623157E+308;
        var d = this.v7(a);
        var e = d.x;
        var f = d.y;
        if (this.yx.c1.e > this.z8) {
            return -1;
        }
        this.yx.c4(function (g) {
            if (g._visibility == 0) {
                var h = _this.yx.cu(g);
                var i = h.left + h.width / 2;
                var j = h.top + h.height / 2;
                var k = (e - i) * (e - i) + (f - j) * (f - j);
                if (k < c) {
                    c = k;
                    b = g;
                }
            }
        });
        if (b != null) {
            return b.ak;
        }
        return -1;
    };
    PolarBase.prototype.wc = function (a, b, c) {
        var _this = this;
        if (this.yt == null || this.yu == null || this.zj == null || this.zk == null) {
            return { $type: Point_$type, x: NaN, y: NaN };
        }
        var d = this.cv.by;
        var e = this.cv.bx;
        var f = this.getEffectiveViewport();
        var g = new ScalerParams(0, d, e, this.yt.ch, f);
        var h = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yu.ch, f);
            $ret.b = _this.iy();
            return $ret;
        })());
        if (b) {
            var i = this.ir(a);
            if (i == -1) {
                return { $type: Point_$type, x: NaN, y: NaN };
            }
            var j = truncate(Math.floor(i));
            var k = truncate(Math.ceil(i));
            if (j < 0) {
                j = 0;
            }
            if (k < 0) {
                k = 0;
            }
            if (k > this.dn.count - 1) {
                k = this.dn.count - 1;
            }
            if (j > this.dn.count - 1) {
                j = this.dn.count - 1;
            }
            if (k == j) {
                return this.yv.h(this.zj.item(j), this.zk.item(j), d, e, f, this.yw.e, this.yw.d, this.yw.g, this.yw.f, this.yw.i, this.yw.h);
            }
            var l = i - Math.floor(i);
            var m = this.yv.h(this.zj.item(j), this.zk.item(j), d, e, f, this.yw.e, this.yw.d, this.yw.g, this.yw.f, this.yw.i, this.yw.h);
            var n = this.yv.h(this.zj.item(k), this.zk.item(k), d, e, f, this.yw.e, this.yw.d, this.yw.g, this.yw.f, this.yw.i, this.yw.h);
            var o = m.x;
            var p = m.y;
            var q = n.x;
            var r = n.y;
            var s = q - o;
            var t = r - p;
            return { $type: Point_$type, x: o + s * l, y: p + t * l };
        }
        else {
            var u = this.j3(a);
            if (u == -1) {
                return { $type: Point_$type, x: NaN, y: NaN };
            }
            return this.yv.h(this.zj.item(u), this.zk.item(u), d, e, f, this.yw.e, this.yw.d, this.yw.g, this.yw.f, this.yw.i, this.yw.h);
        }
    };
    PolarBase.prototype.ai = function () {
        return [this.yt, this.yu];
    };
    PolarBase.prototype.kq = function (a, b) {
        var c = this.nd(b);
        if (c == PolarBase.$$p[3]) {
            return this.kr(a, b, this.aaj);
        }
        if (c == PolarBase.$$p[15]) {
            return this.kr(a, b, this.aa1);
        }
        if (c == PolarBase.$$p[5]) {
            return this.kr(a, b, this.aam);
        }
        if (c == PolarBase.$$p[6]) {
            return this.kr(a, b, this.aao);
        }
        return _super.prototype.kq.call(this, a, b);
    };
    PolarBase.prototype.md = function (a) {
        var b = this.nd(a);
        if (b == PolarBase.$$p[3]) {
            return this.aaj;
        }
        if (b == PolarBase.$$p[15]) {
            return this.aa1;
        }
        if (b == PolarBase.$$p[5]) {
            return this.aam;
        }
        if (b == PolarBase.$$p[6]) {
            return this.aao;
        }
        return _super.prototype.md.call(this, a);
    };
    PolarBase.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        if (this.yx.c0.g$b(a, b, c, d)) {
            this.rr(false);
            this.ql();
        }
        switch (b) {
            case PolarBase.$$p[1]:
            case PolarBase.$$p[2]:
            case PolarBase.$$p[13]:
            case PolarBase.$$p[14]:
                this.qf();
                break;
            case "FastItemsSource":
                if (typeCast(IFastItemsSource_$type, c) != null) {
                    c.deregisterColumn(this.zj);
                    c.deregisterColumn(this.zk);
                    this.zj = null;
                    this.zk = null;
                }
                if (typeCast(IFastItemsSource_$type, d) != null) {
                    this.zj = this.dk(this.aaj);
                    this.zk = this.dk(this.aa1);
                }
                if ((this.yu != null && !this.yu.c1()) || (this.yt != null && !this.yt.c1())) {
                    this.rr(this.gg && c != null);
                }
                break;
            case PolarBase.$$p[0]:
                if (this.yt != null && this.yu != null) {
                    this.yv = new PolarAxes(this.yu, this.yt);
                }
                if (c != null) {
                    c.b0(this);
                }
                if (d != null) {
                    d.ct(this);
                }
                if (this.yt != null && !this.yt.c1()) {
                    this.rr(false);
                }
                else if (c != null && d == null) {
                    this.o5(true, this.cv);
                }
                break;
            case PolarBase.$$p[12]:
                if (this.yt != null && this.yu != null) {
                    this.yv = new PolarAxes(this.yu, this.yt);
                }
                if (c != null) {
                    c.b0(this);
                }
                if (d != null) {
                    d.ct(this);
                }
                if (this.yu != null && !this.yu.c1()) {
                    this.rr(false);
                }
                else if (c != null && d == null) {
                    this.o5(true, this.cv);
                }
                if (this.yt != null && !this.yt.c1()) {
                    this.yt.h7();
                }
                break;
            case PolarBase.$$p[3]:
                if (this.dn != null) {
                    this.dn.deregisterColumn(this.zj);
                    this.zj = this.dk(this.aaj);
                }
                break;
            case "AngleColumn":
                this.yx.c0.g$i();
                if (this.yt != null && !this.yt.c1()) {
                    this.rr(this.gg && this.d7);
                }
                break;
            case PolarBase.$$p[15]:
                if (this.dn != null) {
                    this.dn.deregisterColumn(this.zk);
                    this.zk = this.dk(this.aa1);
                }
                break;
            case "RadiusColumn":
                this.yx.c0.g$i();
                if (this.yu != null && !this.yu.c1()) {
                    this.rr(this.gg && this.d7);
                }
                break;
            case PolarBase.$$p[16]:
                this.rr(false);
                this.ql();
                break;
            case PolarBase.$$p[11]:
                this.rr(false);
                this.ql();
                break;
            case "TransitionProgress":
                this.yw = new PolarAxisInfoCache(this.yt, this.yu, this.dn);
                this.acf = this.cv.by;
                this.ace = this.cv.bx;
                this.acd = this.cv.getEffectiveViewport();
                this.y4.v = this.zz;
                this.y4.c(this.jq, this.y2, this.y0);
                if (this.d5(this.cv)) {
                    return;
                }
                if (this.jq == 1) {
                    this.abh(this.y0, this.yx);
                    this.qo();
                }
                else {
                    this.abh(this.y4, this.yx);
                }
                if (this.db != null) {
                    this.db.qo(this);
                }
                break;
            case "TrendLineBrush":
                this.rf(b);
                break;
            case PolarBase.$$p[4]:
                this.rr(false);
                this.ql();
                break;
            case "TrendLineType":
                this.ql();
                break;
            case PolarBase.$$p[8]:
                this.zm = this.zv();
                this.rr(false);
                break;
            case PolarBase.$$p[7]:
                this.zl = this.zu();
                this.rr(false);
                break;
            case PolarBase.$$p[9]:
                this.y7 = this.y8();
                break;
        }
    };
    Object.defineProperty(PolarBase.prototype, "y7", {
        get: function () {
            return this.y6;
        },
        set: function (a) {
            var b = this.y6;
            this.y6 = a;
            if (b != this.y6) {
                this.raisePropertyChanged("ActualItemSearchMode", enumGetBox(ScatterItemSearchMode_$type, b), enumGetBox(ScatterItemSearchMode_$type, this.y6));
            }
        },
        enumerable: false,
        configurable: true
    });
    PolarBase.prototype.y8 = function () {
        if (this.y9 != 0) {
            return this.y9;
        }
        return this.zt ? 4 : 1;
    };
    PolarBase.prototype.get_zt = function () {
        return false;
    };
    Object.defineProperty(PolarBase.prototype, "zt", {
        get: function () {
            return this.get_zt();
        },
        enumerable: false,
        configurable: true
    });
    PolarBase.prototype.zv = function () {
        return this.zs;
    };
    PolarBase.prototype.zu = function () {
        return this.zr;
    };
    PolarBase.prototype.zo = function (a) {
        if (typeCast(NumericRadiusAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    PolarBase.prototype.zn = function (a) {
        if (typeCast(NumericAngleAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    PolarBase.prototype.o5 = function (a, b) {
        _super.prototype.o5.call(this, a, b);
        var c = b;
        if (a) {
            c.c1.clear();
        }
        c.c0.g$d();
    };
    PolarBase.prototype.yq = function (a, b) {
        var c = a.top - 10;
        var d = a.bottom + 10;
        var e = a.left - 10;
        var f = a.right + 10;
        var g = ((function () {
            var $ret = new Clipper(1, e, d, f, c, false);
            $ret.i = b;
            return $ret;
        })());
        return g;
    };
    PolarBase.prototype.aa5 = function (a, b, c, d) {
        a.g = new Dictionary$2(Base.$, OwnedPoint.$, 1, b);
        var e = this.dn;
        var _loop_1 = function (f) {
            var g = a.w._inner[f];
            if (!isInfinity(g.x) && !isInfinity(g.y)) {
                var h_1 = { $type: Point_$type, x: this_1.zj.item(f), y: this_1.zk.item(f) };
                var i_1 = { $type: Point_$type, x: g.x, y: g.y };
                a.g.addItem(e.item(f), ((function () {
                    var $ret = new OwnedPoint();
                    $ret.a = e.item(f);
                    $ret.b = h_1;
                    $ret.c = i_1;
                    return $ret;
                })()));
            }
        };
        var this_1 = this;
        for (var f = 0; f < b; f++) {
            _loop_1(f);
        }
    };
    PolarBase.prototype.abd = function (a, b) {
        var _this = this;
        var c = b.by;
        var d = b.bx;
        var e = b.getEffectiveViewport();
        a.j.clear();
        a.n.clear();
        var f = Math.min(this.zj != null ? this.zj.count : 0, this.zk != null ? this.zk.count : 0);
        if (f < 1) {
            return;
        }
        this.yw = new PolarAxisInfoCache(this.yt, this.yu, this.dn);
        this.yy = a;
        this.ace = d;
        this.acf = c;
        this.acd = b.getEffectiveViewport();
        b.cy.l(a.j, this.z9, c, d, this.id);
        if (f <= this.z9) {
            this.aa5(a, f, c, d);
        }
        var g = this.yq(d, a.n);
        var h = Math.min(this.yt.l4, this.yt.l3);
        var i = Math.max(this.yt.l3, this.yt.l4);
        b.c0.i = this.zz;
        b.c0.h = 0;
        b.c0.k = this.yu.n7;
        b.c0.j = this.yu.n6;
        b.c0.l = function (j, k) { return _this.yv.c(j, k, c, d, e, Math.cos); };
        b.c0.m = function (j, k) { return _this.yv.d(j, k, c, d, e, Math.sin); };
        var j = new List$1(Number_$type, 0);
        for (var k = 0; k < this.zj.count; k++) {
            if (this.zj.item(k) >= h && this.zj.item(k) <= i) {
                j.add(this.zj.item(k));
            }
        }
        var l = this.zk;
        if (j.count != this.zj.count) {
            l = new List$1(Number_$type, 0);
            for (var m = 0; m < this.zj.count; m++) {
                if (this.zj.item(m) >= h && this.zj.item(m) <= i) {
                    l.add(this.zk.item(m));
                }
            }
        }
        b.c0.n(a.n, this.zb, j, l, this.aaa, runOn(this.yt, this.yt.getScaledAngle), runOn(this.yu, this.yu.od), ((function () {
            var $ret = new TrendResolutionParams();
            $ret.d = _this.id;
            $ret.i = d;
            $ret.h = e;
            $ret.j = c;
            return $ret;
        })()), g, h, i);
    };
    PolarBase.prototype.get_fe = function () {
        return true;
    };
    PolarBase.prototype.zd = function (a, b, c) {
        return null;
    };
    PolarBase.prototype.ze = function (a, b, c) {
        if (this.dn == null) {
            return null;
        }
        var d = this.y1();
        var e = this.yx.b1(this.v7(b));
        var f = -1;
        if (e == null) {
            if (a != null) {
                f = this.dn.indexOf(a);
            }
            else {
                return null;
            }
        }
        else {
            f = e.ak;
        }
        if (f == -1) {
            return null;
        }
        var g = new HighlightingInfo();
        g.i = this;
        g.g = f;
        g.f = f;
        g.c = false;
        g.d = true;
        if (c != null && c.g == g.g && c.f == g.f) {
            return c;
        }
        return g;
    };
    PolarBase.prototype.zc = function (a, b, c) {
        var d = new HighlightingInfo();
        d.i = this;
        d.g = 0;
        d.f = this.dn.count - 1;
        d.c = true;
        if (c != null && c.g == d.g && c.f == d.f) {
            return c;
        }
        return d;
    };
    PolarBase.prototype.dg = function (a, b, c) {
        if (this.ed && this.fz(this.bx) && !c) {
            var d = this.zd(a, b, this.zf);
            this.zf = d;
            return d;
        }
        else {
            var e = this.zc(a, b, this.zf);
            this.zf = e;
            return e;
        }
    };
    PolarBase.prototype.dh = function (a, b, c) {
        if (this.fz(this.bx) && !c) {
            var d = this.ze(a, b, this.zg);
            this.zg = d;
            return d;
        }
        else {
            var e = this.zc(a, b, this.zg);
            e.d = true;
            this.zg = e;
            return e;
        }
    };
    PolarBase.prototype.zx = function () {
        return (this.assigningPolarStyle != null && this.zm) || (this.bx != 1) || this.d2 || (this.cp != 1) || (this.co != 1);
    };
    PolarBase.prototype.zw = function () {
        return (this.assigningPolarMarkerStyle != null && this.zl) || this.d2 || (this.bx != 1) || (this.cp != 1) || (this.co != 1);
    };
    PolarBase.prototype.abh = function (a, b) {
        this.qx();
        var c = b.by;
        var d = b.bx;
        var e = this.getEffectiveViewport1(b);
        var f = this.wj(b);
        this.yw = new PolarAxisInfoCache(this.yt, this.yu, this.dn);
        if (this.xg()) {
            b.cy.j(a.j, this.xi);
            if (this.zw()) {
                this.aa7(b, a);
            }
        }
        b.cr();
        var g = this.yq(d, b.c0.g$j.an);
        b.c0.g$h(a.n, g);
        this.aa4(d, c, e, f, b);
    };
    PolarBase.prototype.yn = function (a, b) {
        return this.yo(a, b);
    };
    PolarBase.prototype.yo = function (a, b) {
        if (b < a || a < 0 || a > this.zj.count || b < 0 || b > this.zj.count) {
            return null;
        }
        var c = new Array((b - a) + 1);
        for (var d = a; d <= b; d++) {
            var e = d;
            c[d - a] = this.dn.item(d);
        }
        return c;
    };
    PolarBase.prototype.aa7 = function (a, b) {
        var _this = this;
        a.b7 = true;
        this.y5.ay(this, this.zw(), runOn(this, this.yn));
        var c = this.y5.d;
        var d = c != null;
        var e = this.getEffectiveViewport1(a);
        var f = this.dn == null ? 0 : this.dn.count;
        this.yx.cj(function (g) {
            if (g._visibility == 0) {
                var h = g.ak;
                var i = g.content;
                if (i == null) {
                    i = new DataContext();
                    g.content = i;
                }
                if (d) {
                    _this.abb(h, f, a.q);
                }
                _this.y5.ac(g, i);
            }
        });
    };
    PolarBase.prototype.abc = function (a, b, c) {
        var d = this.bx != 1;
        var e = this.cp != 1;
        var f = this.co != 1;
        this.y5.a0(this.y5.b, a, b, false);
        this.y5.ab(this.ib);
        var g = this.y5.ax;
        var h = null;
        var i = null;
        var j = null;
        var k = this.dn;
        if (d && this.db != null) {
            h = this.db.dl.getHighlightingInfo(this, k, g.z, g.y, false);
            g.c = h;
        }
        if (e && this.db != null) {
            i = this.db.dm.getHighlightingInfo(this, k, g.z, g.y, false);
            if (i == null && this.bn != null) {
                i = this.db.dm.getHighlightingInfo(this.bn, this.bn.dn, g.z, g.y, false);
            }
            g.d = i;
        }
        if (f && this.db != null) {
            j = this.db.dk.getHighlightingInfo(this, k, g.z, g.y, false);
            g.b = j;
        }
        g.h = c;
        this.abf(g);
        this.y5.z(this, h, i, j, c, d, e, f, a < 0);
    };
    PolarBase.prototype.abb = function (a, b, c) {
        var d = this.bx != 1;
        var e = this.cp != 1;
        var f = this.co != 1;
        this.y5.a0(this.y5.d, a, b, true);
        this.y5.aa(this.ib);
        var g = this.y5.aw;
        var h = null;
        var i = null;
        var j = null;
        var k = this.dn;
        var l = null;
        var m = null;
        var n = null;
        if (d && this.db != null) {
            h = this.db.dl.getHighlightingInfo(this, k, g.z, g.y, true);
            if (this.db.dl.isCrossContextHighlight(this)) {
                l = this.db.dl.getHighlightingInfo(this, k, g.z, g.y, true);
            }
            g.c = h;
        }
        if (e && this.db != null) {
            i = this.db.dm.getHighlightingInfo(this, k, g.z, g.y, true);
            if (i == null && this.bn != null) {
                i = this.db.dm.getHighlightingInfo(this.bn, this.bn.dn, g.z, g.y, true);
            }
            if (this.db.dm.isCrossContextHighlight(this)) {
                m = this.db.dm.getHighlightingInfo(this, k, g.z, g.y, true);
            }
            g.d = i;
        }
        if (f && this.db != null) {
            j = this.db.dk.getHighlightingInfo(this, k, g.z, g.y, true);
            if (this.db.dk.isCrossContextHighlight(this)) {
                n = this.db.dk.getHighlightingInfo(this, k, g.z, g.y, true);
            }
            g.d = i;
        }
        g.h = c;
        this.abe(g);
        this.y5.y(this, h, i, j, c, d, e, f, g.v, this.db.dl.totalHighWaterMark, g.x, this.db.dm.totalHighWaterMark, g.t, this.db.dk.totalHighWaterMark);
    };
    PolarBase.prototype.abf = function (a) {
        if (this.assigningPolarStyle != null && this.zm) {
            this.assigningPolarStyle(this, a);
        }
    };
    PolarBase.prototype.abe = function (a) {
        if (this.assigningPolarMarkerStyle != null && this.zl) {
            this.assigningPolarMarkerStyle(this, a);
        }
    };
    PolarBase.prototype.aa4 = function (a, b, c, d, e) {
        e.c3(a, b, c, d);
    };
    PolarBase.prototype.sh = function () {
        _super.prototype.sh.call(this);
        if (this.index < 0) {
            return;
        }
        this.yx.c6();
    };
    PolarBase.prototype.rt = function (a, b) {
        _super.prototype.rt.call(this, a, b);
        if (!this.g0) {
            this.cv.a4(b);
            return;
        }
        this.cv.a4(b);
        if (this.d5(this.cu)) {
            return;
        }
        var c = this.cu;
        if (!this.gs) {
            this.y3 = new PolarFrame();
            this.abd(this.y3, c);
        }
        this.gs = false;
        this.abh(this.y3, c);
        this.g0 = false;
    };
    PolarBase.prototype.rq = function (a, b, c, d, e) {
        _super.prototype.rq.call(this, a, b, c, d, e);
        var f = this.ba.a$j.item(d);
        var g = f;
        f.a3(c);
        if (this.d5(g)) {
            return;
        }
        if (this.yz == null) {
            this.yz = new PolarFrame();
        }
        this.abd(this.yz, g);
        this.abh(this.yz, g);
    };
    PolarBase.prototype.abg = function (a, b) {
        var e_1, _a, e_2, _b;
        var c = new List$1(Base.$, 0);
        try {
            for (var _c = __values(fromEnum(b.c)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var d = _d.value;
                if (!a.containsKey(d)) {
                    c.add1(d);
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        try {
            for (var _e = __values(fromEnum(c)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var e = _f.value;
                b.h(e);
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
    };
    PolarBase.prototype.ym = function (a, b, c) {
        this.aa8();
        return this.yy.w.toArray();
    };
    PolarBase.prototype.z1 = function (a) {
        var e_3, _a;
        var b = new List$1(Number_$type, 0);
        var c = this.dn;
        try {
            for (var _b = __values(fromEnum(this.yx.c1.c)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var d = _c.value;
                b.add(c.indexOf(d));
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
        return b;
    };
    PolarBase.prototype.acb = function (a) {
        return { $type: Point_$type, x: this.zj.item(a), y: this.zk.item(a) };
    };
    PolarBase.prototype.pz = function (a) {
        _super.prototype.pz.call(this, a);
        this.bc.exportTrendlineData(this, a, this.yx.c0.g$j);
    };
    PolarBase.prototype.getDataLegendSeriesContextAt = function (a, b) {
        var c = { $type: Point_$type, x: a, y: b };
        var d = this.kn(c);
        var e = d == null ? NaN : this.ix(d, "RadiusMemberPath");
        var f = d == null ? NaN : this.ix(d, "AngleMemberPath");
        var g = new DataLegendSeriesValueInfo();
        g.l = e;
        g.a = 11;
        g.r = this.aa1 != null ? this.aa1 : "Radius";
        g.q = "Radius";
        g.s = this.aax != null ? this.aax : "R:";
        g.t = this.aaz != null ? this.aaz : "";
        var h = new DataLegendSeriesValueInfo();
        h.l = f;
        h.a = 10;
        h.r = this.aaj != null ? this.aaj : "Angle";
        h.q = "Angle";
        h.s = this.aaf != null ? this.aaf : "A:";
        h.t = this.aah != null ? this.aah : "";
        var i = new DataLegendSeriesContext();
        i.e = ((function () {
            var $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
            $ret.add(g);
            $ret.add(h);
            return $ret;
        })());
        i.h = this.mc();
        return i;
    };
    PolarBase.prototype.ak = function (a, b) {
        var c = new List$1(ChartSelection.$, 0);
        this.p4(a, c, b);
        return c.toArray();
    };
    PolarBase.prototype.p4 = function (a, b, c) {
        var _this = this;
        var d = this.cv.by;
        var e = this.cv.bx;
        var f = this.getEffectiveViewport1(this.cv);
        var g = this.wj(this.cv);
        var h = new CollisionRect(0, a);
        var i = this.yu.k5(e, d, f, g);
        var _loop_2 = function (j) {
            var k = this_2.yt.getScaledAngle(this_2.zj.item(j));
            var l = this_2.yu.od(this_2.zk.item(j));
            var m = this_2.yv.c(k, l, d, e, f, Math.cos);
            var n = this_2.yv.d(k, l, d, e, f, Math.sin);
            if (h.containsLocation(m, n)) {
                if (c == 7 || c == 6) {
                    b.add(((function () {
                        var $ret = new ChartSelection();
                        $ret.series = _this;
                        return $ret;
                    })()));
                    return { value: void 0 };
                }
                else if (c == 2 || c == 1) {
                    b.add(((function () {
                        var $ret = new ChartSelection();
                        $ret.item = _this.dn.item(j);
                        return $ret;
                    })()));
                }
                else {
                    b.add(((function () {
                        var $ret = new ChartSelection();
                        $ret.item = _this.dn.item(j);
                        $ret.series = _this;
                        return $ret;
                    })()));
                }
            }
        };
        var this_2 = this;
        for (var j = 0; j < this.dn.count; j++) {
            var state_1 = _loop_2(j);
            if (typeof state_1 === "object")
                return state_1.value;
        }
    };
    PolarBase.$t = markType(PolarBase, 'PolarBase', MarkerSeries.$);
    PolarBase.abi = DependencyProperty.i("ActualTrendLineBrush", Brush.$, PolarBase.$, new PropertyMetadata(2, null, function (a, b) { return a.raisePropertyChanged("ActualTrendLineBrush", b.oldValue, b.newValue); }));
    PolarBase.$$p = markDep(DependencyProperty, PropertyMetadata, PolarBase, 'raisePropertyChanged', ['AngleAxis:yt:abj', [NumericAngleAxis.$, null], 'AngleMemberAsLegendLabel:aaf:abk', [2, null], 'AngleMemberAsLegendUnit:aah:abl', [2, null], 'AngleMemberPath:aaj:abm', [2], 'ClipSeriesToBounds:zp:abn', [0, false], 'HighlightedAngleMemberPath:aam:abo', [2, null], 'HighlightedRadiusMemberPath:aao:abp', [2, null], 'IsCustomPolarMarkerStyleAllowed:zr:abq', [0, false], 'IsCustomPolarStyleAllowed:zs:abr', [0, false], 'ItemSearchMode:y9:abs', [ScatterItemSearchMode_$type, enumGetBox(ScatterItemSearchMode_$type, 0)], 'ItemSearchThreshold:z8:abt', [1, 10000], 'MaximumMarkers:z9:abu', [1, 400], 'RadiusAxis:yu:abv', [NumericRadiusAxis.$, null], 'RadiusMemberAsLegendLabel:aax:abw', [2, null], 'RadiusMemberAsLegendUnit:aaz:abx', [2, null], 'RadiusMemberPath:aa1:aby', [2], 'UseCartesianInterpolation:zz:ab6', [0, true], 'TrendLineType:zb:ab4', [TrendLineType_$type, enumGetBox(TrendLineType_$type, 0)], 'TrendLineThickness:z7:ab3', [1, 1.5], 'TrendLinePeriod:aaa:ab2', [1, 7], 'TrendLineZIndex:aab:ab5', [1, 1], 'TrendLineBrush:ab8:abz', [Brush.$, null], 'TrendLineDashArray:ab9:ab0', [DoubleCollection.$], 'TrendLineDashCap:aca:ab1', [PenLineCap_$type, enumGetBox(PenLineCap_$type, 0)]]);
    return PolarBase;
}(MarkerSeries));
export { PolarBase };
/**
 * @hidden
 */
var PolarSeriesRenderManager = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarSeriesRenderManager, _super);
    function PolarSeriesRenderManager() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.aw = null;
        _this.ax = null;
        return _this;
    }
    PolarSeriesRenderManager.prototype.c = function () {
        this.aw = new AssigningPolarMarkerStyleEventArgs();
        return this.aw;
    };
    PolarSeriesRenderManager.prototype.a = function () {
        this.ax = new AssigningPolarStyleEventArgs();
        return this.ax;
    };
    PolarSeriesRenderManager.prototype.ay = function (a, b, c) {
        this.w(a, b, c);
    };
    PolarSeriesRenderManager.prototype.az = function (a, b, c) {
        this.x(a, b, c);
    };
    PolarSeriesRenderManager.prototype.a0 = function (a, b, c, d) {
        if (b == -1) {
            a.e = false;
            a.z = 0;
            a.y = c - 1;
        }
        else {
            if (d) {
                a.z = b;
                a.y = b;
            }
            else {
                a.e = false;
                a.z = b;
                a.y = b;
            }
        }
    };
    PolarSeriesRenderManager.$t = markType(PolarSeriesRenderManager, 'PolarSeriesRenderManager', SeriesRenderManager.$);
    return PolarSeriesRenderManager;
}(SeriesRenderManager));
export { PolarSeriesRenderManager };
/**
 * @hidden
 */
var PolarBaseView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarBaseView, _super);
    function PolarBaseView(a) {
        var _this = _super.call(this, a) || this;
        _this.cz = null;
        _this.cy = null;
        _this.c0 = null;
        _this.c1 = null;
        _this.cz = a;
        _this.c1 = new HashPool$2(Base.$, Marker.$);
        _this.cm(_this.c1);
        _this.c0 = new PolarTrendLineManager();
        return _this;
    }
    PolarBaseView.prototype.au = function () {
        _super.prototype.au.call(this);
        this.cy = this.cx();
        if (!this.q) {
            this.cz.z9 = 1000;
            this.b4.xa = Defaults.e;
        }
    };
    PolarBaseView.prototype.c4 = function (a) {
        this.c1.g(a);
    };
    PolarBaseView.prototype.cx = function () {
        var _this = this;
        var a = new NumericMarkerManager(0, function (b) { return _this.c1.item(b); }, function (b) { return _this.cz.yw.c.item(b); }, runOn(this, this.c5), runOn(this, this.cw), runOn(this, this.c2));
        a.n = true;
        a.p = runOn(this.cz, this.cz.acb);
        return a;
    };
    PolarBaseView.prototype.c5 = function (a) {
        this.cz.abg(a, this.c1);
    };
    PolarBaseView.prototype.cw = function () {
        return this.cz.ym(this.c1, this.by, this.bx);
    };
    PolarBaseView.prototype.c2 = function () {
        return this.cz.z1(this.c1);
    };
    PolarBaseView.prototype.c3 = function (a, b, c, d) {
        if (this.cz.zp) {
            var e = new GeometryGroup();
            this.cz.yu.ol(e, a, b, c, d);
        }
        else {
        }
    };
    PolarBaseView.prototype.c6 = function () {
        if (this.cz.ab8 != null) {
            this.cz.ab7 = this.cz.ab8;
        }
        else {
            this.cz.ab7 = this.cz.u9;
        }
    };
    PolarBaseView.prototype.cj = function (a) {
        this.c1.f(a);
    };
    PolarBaseView.prototype.a7 = function (a, b) {
        if (a.d) {
            if (this.cz.zb != 0 && !b) {
                var c = this.c0.g$j;
                c.ad = this.cz.z7;
                c._stroke = this.cz.ab7;
                c.aj = this.cz.aca;
                c.ai = this.cz.ab9;
                a.v(c);
            }
        }
        _super.prototype.a7.call(this, a, b);
    };
    PolarBaseView.prototype.c7 = function () {
        return Rect.empty;
    };
    PolarBaseView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span>" + "${item." + this.cz.aaj + "}" + "</span><br/><span";
        var b = this.e.ba.a$u(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.cz.title + ": </span><span class='ui-priority-primary'>" + "${item." + this.cz.aa1 + "}</span></div>";
        return a;
    };
    PolarBaseView.$t = markType(PolarBaseView, 'PolarBaseView', MarkerSeriesView.$);
    return PolarBaseView;
}(MarkerSeriesView));
export { PolarBaseView };
