// igniteui-react-charts/esm5/lib/PriceChannelOverlay_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { FinancialOverlay } from "./FinancialOverlay";
import { DependencyProperty } from "igniteui-react-core";
import { Series } from "./Series";
import { List$1 } from "igniteui-react-core";
import { Number_$type, typeCast, runOn, markType, markDep, Point_$type, fromEnum } from "igniteui-react-core";
import { CategoryFrame } from "./CategoryFrame";
import { MathUtil } from "igniteui-react-core";
import { FinancialSeriesView } from "./FinancialSeriesView";
import { ScalerParams } from "./ScalerParams";
import { ISortingAxis_$type } from "./ISortingAxis";
import { FinancialBucketCalculator } from "./FinancialBucketCalculator";
import { Polyline } from "igniteui-react-core";
import { Polygon } from "igniteui-react-core";
import { RangeValueList } from "./RangeValueList";
import { PropertyMetadata } from "igniteui-react-core";
import { truncate, isNaN_, isInfinity } from "igniteui-react-core";
import { Flattener } from "igniteui-react-core";
/**
 * @hidden
 */
var PriceChannelOverlay = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PriceChannelOverlay, _super);
    function PriceChannelOverlay() {
        var _this = _super.call(this) || this;
        _this.aa7 = new List$1(Number_$type, 0);
        _this.aa6 = new List$1(Number_$type, 0);
        _this.ab = PriceChannelOverlay.$;
        _this.xa = new CategoryFrame(3);
        _this.xc = new CategoryFrame(3);
        _this.w8 = new CategoryFrame(3);
        return _this;
    }
    PriceChannelOverlay.prototype.bj = function () {
        return new PriceChannelOverlay();
    };
    PriceChannelOverlay.prototype.q9 = function (a, b, c, d) {
        switch (b) {
            case "HighColumn":
            case "LowColumn":
                this.aa0 = false;
                break;
        }
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case PriceChannelOverlay.$$p[0]:
                this.aa0 = false;
                this.rr(false);
                break;
        }
    };
    PriceChannelOverlay.prototype.he = function (a, b, c) {
        var d = _super.prototype.he.call(this, a, b, c);
        if (this.x0 == null || this.x1 == null) {
            d = false;
        }
        return d;
    };
    PriceChannelOverlay.prototype.aa1 = function () {
        this.aa7.clear();
        this.aa6.clear();
        var a = truncate(MathUtil.d(this.aa8, 0, this.dn.count));
        var b = Math.min(this.x0.count, this.x1.count);
        var c = this.yq(this.x0);
        var d = this.yq(this.x1);
        for (var e = 0; e < b; e++) {
            var f = Math.min(a, e);
            var g = -1.7976931348623157E+308;
            var h = 1.7976931348623157E+308;
            for (var i = 0; i < f; i++) {
                if (!isNaN_(c.item(e - i))) {
                    g = Math.max(g, c.item(e - i));
                }
                if (!isNaN_(d.item(e - i))) {
                    h = Math.min(h, d.item(e - i));
                }
            }
            if (e == 0) {
                h = d.item(0);
                g = c.item(0);
            }
            this.aa7.add(g);
            this.aa6.add(h);
        }
        return true;
    };
    PriceChannelOverlay.prototype.aac = function (a, b) {
        _super.prototype.aac.call(this, a, b);
        var c = b.by;
        var d = b.bx;
        var e = this.getEffectiveViewport1(b);
        var f = new ScalerParams(0, c, d, this.w6.ch, e);
        var g = new ScalerParams(0, c, d, this.xp.ch, e);
        g.b = this.iy();
        var h = this.w6;
        var i = this.xp;
        a.f.clear();
        a.n.clear();
        a.q.clear();
        var j = 0;
        var k = typeCast(ISortingAxis_$type, this.w6);
        if (k != null && k.i$b.count != this.dn.count) {
            return;
        }
        j = this.yx(c, d, e);
        if (!this.aa0) {
            this.aa0 = this.aa1();
        }
        g.b = this.iy();
        var l = this.w6.eo(2, f) - this.w6.eo(1, f);
        for (var m = b.b0.c; m <= b.b0.d; ++m) {
            var n = void 0;
            if (k == null) {
                n = b.b0.getBucket(m);
            }
            else {
                var o = k.i$b._inner[m];
                var p = k.i$d(o);
                var q = this.aa7._inner[m];
                var r = this.aa6._inner[m];
                var s = q;
                var t = r;
                var u = p;
                while (m < b.b0.d) {
                    o = k.i$b._inner[m + 1];
                    u = k.i$d(o);
                    if (u - p > l) {
                        break;
                    }
                    m++;
                    s = Math.max(q, this.aa7._inner[m]);
                    t = Math.min(r, this.aa6._inner[m]);
                }
                if (!isInfinity(t) && !isInfinity(s)) {
                    var v = NaN;
                    if (!isNaN_(p)) {
                        v = this.w6.el(p, f);
                    }
                    n = [v, t, s];
                }
                else {
                    n = [NaN, NaN, NaN];
                }
            }
            var w = Math.max(1, l);
            if (!isNaN_(n[0]) && m * w >= this.aa2) {
                if (this.w6 != null && typeCast(ISortingAxis_$type, this.w6) !== null) {
                    n[0] = n[0] + j;
                }
                else {
                    n[0] = h.el(n[0], f) + j;
                }
                n[1] = i.el(n[1], g);
                n[2] = i.el(n[2], g);
                a.f.add(n);
            }
        }
    };
    PriceChannelOverlay.prototype.o5 = function (a, b) {
        _super.prototype.o5.call(this, a, b);
        var c = b;
        if (c != null) {
            c.b7();
        }
    };
    PriceChannelOverlay.prototype.aae = function (a, b) {
        _super.prototype.aae.call(this, a, b);
        var c = typeCast(PriceChannelOverlayView.$, b);
        if (c == null) {
            return;
        }
        var d = this.getEffectiveViewport1(b);
        if (b.b3(a)) {
            c.b7();
            var e_1 = a.f.count;
            var f = function (g) { return a.f._inner[g][0]; };
            var g = function (h) { return a.f._inner[e_1 - 1 - h][0]; };
            var h = function (i) { return a.f._inner[i][1]; };
            var i = function (j) { return a.f._inner[e_1 - 1 - j][2]; };
            c.b8(e_1, f, g, h, i);
            b.b5(a);
        }
        this.xf.a2(this, this.ye(), runOn(this, this.wy), this.y3(b), this.y5(b));
        var j = false;
        var k = this.xf.b;
        var l = a.f;
        var m = this.dn.count;
        if (k != null) {
            j = true;
        }
        if (j) {
            var n = new ScalerParams(0, b.by, b.bx, this.w6.ch, d);
            this.aab(l, -1, m, this.w6, n, b.q, -1);
        }
        var o = c.ce;
        var p = c.cf;
        var q = c.ca;
        this.xf.ad(o, true, false, true, true);
        this.xf.ad(p, true, false, true, true);
        this.xf.ad(q, false, true, false, false);
        q._opacity = this.xf.j * this.h9;
    };
    PriceChannelOverlay.prototype.cs = function () {
        return new PriceChannelOverlayView(this);
    };
    PriceChannelOverlay.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.aa5 = typeCast(PriceChannelOverlayView.$, a);
    };
    PriceChannelOverlay.prototype.i5 = function (a, b, c) {
        if (this.db == null) {
            return NaN;
        }
        var d = this.getEffectiveViewport1(this.cv);
        var e = new ScalerParams(0, this.db.v4, this.cv.bx, this.w6.ch, d);
        var f = this.yx(this.db.v4, this.cv.bx, d);
        return this.i7(new RangeValueList(this.aa7, this.aa6), a, this.w6, e, f, runOn(this, this.yw), b, c);
    };
    PriceChannelOverlay.prototype.j8 = function (a, b) {
        if (this.aa7 == null || this.aa6 == null) {
            return _super.prototype.j8.call(this, a, b);
        }
        return this.j9(a, b, this.w6, runOn(this, this.yw), new RangeValueList(this.aa7, this.aa6));
    };
    PriceChannelOverlay.prototype.ka = function (a, b) {
        if (this.aa7 == null || this.aa6 == null) {
            return _super.prototype.ka.call(this, a, b);
        }
        return this.kb(a, b, this.w6, runOn(this, this.yw), new RangeValueList(this.aa7, this.aa6));
    };
    PriceChannelOverlay.prototype.iy = function () {
        return Series.i0(this.aa6, null);
    };
    PriceChannelOverlay.$t = markType(PriceChannelOverlay, 'PriceChannelOverlay', FinancialOverlay.$);
    PriceChannelOverlay.$$p = markDep(DependencyProperty, PropertyMetadata, PriceChannelOverlay, 'raisePropertyChanged', ['Period:aa8:aba', [1, 14]]);
    return PriceChannelOverlay;
}(FinancialOverlay));
export { PriceChannelOverlay };
/**
 * @hidden
 */
var PriceChannelBucketCalculator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PriceChannelBucketCalculator, _super);
    function PriceChannelBucketCalculator(a) {
        var _this = _super.call(this, a) || this;
        _this.f = null;
        _this.f = a;
        return _this;
    }
    PriceChannelBucketCalculator.prototype.getBucket = function (a) {
        var b = a * this.b;
        var c = Math.min(b + this.b - 1, this.a.b2.dn.count - 1);
        if (b <= c) {
            var d = -1.7976931348623157E+308;
            var e = 1.7976931348623157E+308;
            var f = 0;
            for (var g = b; g <= c; ++g) {
                if (!isNaN_(this.f.b6.aa7._inner[g]) && !isNaN_(this.f.b6.aa6._inner[g])) {
                    d = Math.max(d, this.f.b6.aa7._inner[g]);
                    e = Math.min(e, this.f.b6.aa6._inner[g]);
                    ++f;
                }
            }
            if (f > 0 && e != 1.7976931348623157E+308 && d != -1.7976931348623157E+308) {
                return [0.5 * (b + c), e, d];
            }
        }
        return [NaN, NaN, NaN];
    };
    PriceChannelBucketCalculator.$t = markType(PriceChannelBucketCalculator, 'PriceChannelBucketCalculator', FinancialBucketCalculator.$);
    return PriceChannelBucketCalculator;
}(FinancialBucketCalculator));
export { PriceChannelBucketCalculator };
/**
 * @hidden
 */
var PriceChannelOverlayView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PriceChannelOverlayView, _super);
    function PriceChannelOverlayView(a) {
        var _this = _super.call(this, a) || this;
        _this.cb = new Polygon();
        _this.cg = new Polyline();
        _this.ch = new Polyline();
        _this.b6 = null;
        _this.b9 = new Polygon();
        _this.cc = new Polyline();
        _this.cd = new Polyline();
        _this.b6 = a;
        return _this;
    }
    Object.defineProperty(PriceChannelOverlayView.prototype, "ca", {
        get: function () {
            return this.cb;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PriceChannelOverlayView.prototype, "ce", {
        get: function () {
            return this.cg;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PriceChannelOverlayView.prototype, "cf", {
        get: function () {
            return this.ch;
        },
        enumerable: false,
        configurable: true
    });
    PriceChannelOverlayView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
        }
    };
    PriceChannelOverlayView.prototype.b1 = function () {
        return new PriceChannelBucketCalculator(this);
    };
    PriceChannelOverlayView.prototype.b7 = function () {
        this.cb.an.clear();
        this.cg.an.clear();
        this.ch.an.clear();
        this.as();
    };
    PriceChannelOverlayView.prototype.b8 = function (a, b, c, d, e) {
        var e_2, _a, e_3, _b;
        try {
            for (var _c = __values(fromEnum(Flattener.d(a, b, d, this.e.id))), _d = _c.next(); !_d.done; _d = _c.next()) {
                var f = _d.value;
                this.cb.an.add({ $type: Point_$type, x: b(f), y: d(f) });
                this.cg.an.add({ $type: Point_$type, x: b(f), y: d(f) });
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
        try {
            for (var _e = __values(fromEnum(Flattener.d(a, c, e, this.e.id))), _f = _e.next(); !_f.done; _f = _e.next()) {
                var g = _f.value;
                this.cb.an.add({ $type: Point_$type, x: c(g), y: e(g) });
                this.ch.an.add({ $type: Point_$type, x: c(g), y: e(g) });
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
        this.as();
    };
    PriceChannelOverlayView.prototype.be = function () {
        _super.prototype.be.call(this);
        this.b9.an = this.cb.an;
        this.cc.an = this.cg.an;
        this.cd.an = this.ch.an;
        var a = this.br();
        this.b9._fill = a;
        this.b9._opacity = 1;
        this.cc._stroke = a;
        this.cc.ad = this.e.ie + 3;
        this.cd._stroke = a;
        this.cd.ad = this.e.ie + 3;
    };
    PriceChannelOverlayView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (a.d) {
            if (b) {
                a.u(this.b9);
                a.v(this.cc);
                a.v(this.cd);
            }
            else {
                a.u(this.cb);
                a.v(this.cg);
                a.v(this.ch);
            }
        }
    };
    PriceChannelOverlayView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        this.e.bc.exportPolygonData(a, this.cb, "fill", ["Fill"]);
        this.e.bc.exportPolylineData(a, this.cg, "bottom", ["Lower", "Main"]);
        this.e.bc.exportPolylineData(a, this.ch, "top", ["Upper"]);
    };
    PriceChannelOverlayView.$t = markType(PriceChannelOverlayView, 'PriceChannelOverlayView', FinancialSeriesView.$);
    return PriceChannelOverlayView;
}(FinancialSeriesView));
export { PriceChannelOverlayView };
