// igniteui-react-charts/esm5/lib/RadialSeriesRenderManager.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { SeriesRenderManager } from "./SeriesRenderManager";
import { AssigningRadialMarkerStyleEventArgs } from "./AssigningRadialMarkerStyleEventArgs";
import { AssigningRadialStyleEventArgs } from "./AssigningRadialStyleEventArgs";
import { markType } from "igniteui-react-core";
import { intDivide } from "igniteui-react-core";
/**
 * @hidden
 */
var RadialSeriesRenderManager = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RadialSeriesRenderManager, _super);
    function RadialSeriesRenderManager() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.az = 0;
        _this.a0 = 0;
        _this.ax = null;
        _this.ay = null;
        return _this;
    }
    RadialSeriesRenderManager.prototype.c = function () {
        this.ax = new AssigningRadialMarkerStyleEventArgs();
        return this.ax;
    };
    RadialSeriesRenderManager.prototype.a = function () {
        this.ay = new AssigningRadialStyleEventArgs();
        return this.ay;
    };
    RadialSeriesRenderManager.prototype.a1 = function (a, b, c, d, e) {
        this.az = d;
        this.a0 = e;
        this.w(a, b, c);
    };
    RadialSeriesRenderManager.prototype.a2 = function (a, b, c, d, e) {
        this.az = d;
        this.a0 = e;
        this.x(a, b, c);
    };
    RadialSeriesRenderManager.prototype.aw = function (a, b) {
        var c = this.az;
        var d = Math.min(b * c, a - 1);
        var e = Math.min(d + c - 1, a - 1);
        var f = this.a0;
        if (((d + this.a0) * c) > a - 1) {
            d = intDivide((((d + this.a0) * c) - a), c);
            e = Math.min(d + c - 1, a - 1);
            f = 0;
        }
        var g = new Array(2);
        g[0] = Math.min(d + f * c, a - 1);
        g[1] = Math.min(e + f * c, a - 1);
        return g;
    };
    RadialSeriesRenderManager.prototype.a3 = function (a, b, c, d, e, f, g, h, i) {
        if (d == -1) {
            {
                a.e = false;
                a.z = 0;
                a.y = e - 1;
            }
        }
        else {
            {
                if (i) {
                    a.z = d;
                    a.y = d;
                }
                else {
                    var j = this.aw(e, d);
                    a.e = false;
                    a.z = j[0];
                    a.y = j[1];
                }
            }
        }
    };
    RadialSeriesRenderManager.$t = markType(RadialSeriesRenderManager, 'RadialSeriesRenderManager', SeriesRenderManager.$);
    return RadialSeriesRenderManager;
}(SeriesRenderManager));
export { RadialSeriesRenderManager };
