// igniteui-react-charts/esm5/lib/RangeColumnSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { HorizontalRangeCategorySeries } from "./HorizontalRangeCategorySeries";
import { DependencyProperty } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { typeCast, runOn, markType, markDep, fromEnum } from "igniteui-react-core";
import { Rect } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { List$1 } from "igniteui-react-core";
import { Pool$1 } from "igniteui-react-core";
import { Rectangle } from "igniteui-react-core";
import { ChartSelection } from "./ChartSelection";
import { PropertyMetadata } from "igniteui-react-core";
import { truncate, isNaN_, isInfinity } from "igniteui-react-core";
import { RangeCategorySeriesView } from "./RangeCategorySeriesView";
/**
 * @hidden
 */
var RangeColumnSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RangeColumnSeries, _super);
    function RangeColumnSeries() {
        var _this = _super.call(this) || this;
        _this.ab = RangeColumnSeries.$;
        return _this;
    }
    RangeColumnSeries.prototype.bj = function () {
        return new RangeColumnSeries();
    };
    RangeColumnSeries.prototype.get_ez = function () {
        return true;
    };
    RangeColumnSeries.prototype.get_fl = function () {
        return true;
    };
    RangeColumnSeries.prototype.jk = function () {
        if (!isNaN_(this.jj)) {
            return _super.prototype.jk.call(this);
        }
        return Defaults.h;
    };
    RangeColumnSeries.prototype.cs = function () {
        return new RangeColumnSeriesView(this);
    };
    RangeColumnSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.ab0 = a;
    };
    RangeColumnSeries.prototype.get_ed = function () {
        return true;
    };
    RangeColumnSeries.prototype.preferredCategoryMode = function (a) {
        return 2;
    };
    RangeColumnSeries.prototype.wn = function (a) {
        if (this.xAxis == null || this.yAxis == null) {
            return Rect.empty;
        }
        var b = this.v7(a);
        var c = this.cv.by;
        var d = this.cv.bx;
        var e = this.getEffectiveViewport1(this.cv);
        var f = new ScalerParams(0, c, d, this.yAxis.ch, e);
        f.b = this.iy();
        var g = this.an(this.ys(), this.yu.f, this.zv(this.cv), this.zw(this.cv), a, true);
        if (g == null) {
            return Rect.empty;
        }
        var h = g[0];
        var i = g[1];
        var j = Math.abs(b.x - h[0]);
        var k = Math.abs(b.x - i[0]);
        var l = this.xAxis.k8(c, d, e);
        if (j < k) {
            var m = h[0] - 0.5 * l;
            var n = h[1];
            var o = h[2];
            return new Rect(0, m, Math.min(n, o), l, Math.max(n, o) - Math.min(n, o));
        }
        else {
            var p = i[0] - 0.5 * l;
            var q = i[1];
            var r = i[2];
            return new Rect(0, p, Math.min(q, r), l, Math.max(q, r) - Math.min(q, r));
        }
    };
    RangeColumnSeries.prototype.gw = function (a, b) {
        if (this.gz(a, b)) {
            return true;
        }
        if (this.xh(a, b)) {
            return true;
        }
        return false;
    };
    RangeColumnSeries.prototype.o5 = function (a, b) {
        _super.prototype.o5.call(this, a, b);
        var c = b;
        if (a && c.c8 != null) {
            c.c8.count = 0;
        }
    };
    RangeColumnSeries.prototype.iw = function () {
        return this.xAxis.k8(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
    };
    RangeColumnSeries.prototype.aak = function (a, b) {
        _super.prototype.aak.call(this, a, b);
        var c = typeCast(RangeColumnSeriesView.$, b);
        var d = a.f;
        var e = b.by;
        var f = b.bx;
        var g = this.getEffectiveViewport1(b);
        var h = this.xAxis.k8(e, f, g);
        if (isNaN_(h) || isInfinity(h)) {
            c.c8.count = 0;
            return;
        }
        this.y2.a2(this, this.zm(), runOn(this, this.ym), this.zv(b), this.zw(b));
        this.y2.r = this.ab1;
        this.y2.s = this.ab2;
        this.y2.k = this.ab1;
        this.y2.l = this.ab2;
        var i = false;
        var j = this.y2.b;
        if (j != null) {
            i = true;
        }
        var k = this.xAxis.cn;
        var l = this.lowColumn.count;
        var m = this.xAxis;
        var n = new ScalerParams(0, e, f, this.xAxis.ch, g);
        var o = 0;
        for (var p = 0; p < d.count; ++p) {
            var q = d._inner[p][0] - 0.5 * h;
            var r = Math.min(d._inner[p][1], d._inner[p][2]);
            var s = Math.max(d._inner[p][1], d._inner[p][2]);
            var t = Math.abs(s - r);
            if (isInfinity(t) || isNaN_(t)) {
                continue;
            }
            var u = c.c8.item(o);
            o++;
            u.width = h;
            u.height = t;
            var v = this.y2.aw(l, p)[0];
            if (k && a.m != null && p >= 0 && p < a.m.count) {
                v = truncate(Math.round(a.m._inner[p]));
            }
            if (i) {
                this.aag(d, p, l, m, n, b.q, v);
            }
            this.y2.ad(u, false, false, false, false);
            u.ap = this.y2.k;
            u.aq = this.y2.l;
            c.dd(u, q, r);
        }
        c.c8.count = o;
        b.c3(a);
    };
    RangeColumnSeries.prototype.p4 = function (a, b, c) {
        this.aac(a, b, c, this.yt, 1);
    };
    RangeColumnSeries.prototype.ak = function (a, b) {
        var c = new List$1(ChartSelection.$, 0);
        this.p4(a, c, b);
        return c.toArray();
    };
    Object.defineProperty(RangeColumnSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zq;
        },
        enumerable: false,
        configurable: true
    });
    RangeColumnSeries.$t = markType(RangeColumnSeries, 'RangeColumnSeries', HorizontalRangeCategorySeries.$);
    RangeColumnSeries.$$p = markDep(DependencyProperty, PropertyMetadata, RangeColumnSeries, 'raisePropertyChanged', ['RadiusX:ab1:ab5', [1, 0], 'RadiusY:ab2:ab6', [1, 0]]);
    return RangeColumnSeries;
}(HorizontalRangeCategorySeries));
export { RangeColumnSeries };
/**
 * @hidden
 */
var RangeColumnSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RangeColumnSeriesView, _super);
    function RangeColumnSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c7 = null;
        _this.c8 = null;
        _this.c9 = null;
        _this.de = new Rectangle();
        _this.c7 = a;
        _this.c8 = ((function () {
            var $ret = new Pool$1(Rectangle.$);
            $ret.create = runOn(_this, _this.df);
            $ret.activate = runOn(_this, _this.da);
            $ret.disactivate = runOn(_this, _this.dc);
            $ret.destroy = runOn(_this, _this.db);
            return $ret;
        })());
        return _this;
    }
    RangeColumnSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        this.c9 = new List$1(Rectangle.$, 0);
        if (!this.q) {
            this.c7.xa = Defaults.c;
            this.c7.jp = Defaults.i;
        }
    };
    RangeColumnSeriesView.prototype.df = function () {
        var a = new Rectangle();
        this.c9.add(a);
        a._visibility = 1;
        return a;
    };
    RangeColumnSeriesView.prototype.da = function (a) {
        a._visibility = 0;
    };
    RangeColumnSeriesView.prototype.dc = function (a) {
        a._visibility = 1;
    };
    RangeColumnSeriesView.prototype.db = function (a) {
        this.c9.remove(a);
    };
    RangeColumnSeriesView.prototype.dd = function (a, b, c) {
        if (!this.isDirty) {
            this.as();
        }
        a.o = c;
        a.n = b;
    };
    RangeColumnSeriesView.prototype.ac = function (a) {
        return this.c9._inner[a];
    };
    RangeColumnSeriesView.prototype.ab = function (a) {
        var b = this.c9._inner[a];
        this.de._visibility = b._visibility;
        this.de.n = b.n;
        this.de.o = b.o;
        this.de.width = b.width;
        this.de.height = b.height;
        var c = this.bs(a);
        this.de._fill = c;
        this.de._stroke = c;
        this.de.ad = this.e.ie + 3;
        return this.de;
    };
    RangeColumnSeriesView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (a.d) {
            for (var c = 0; c < this.c9.count; c++) {
                var d = this.aa(c, b);
                this.bf(d, c, b);
                a.w(d);
            }
        }
    };
    RangeColumnSeriesView.prototype.an = function (a) {
        var e_1, _a, e_2, _b;
        _super.prototype.an.call(this, a);
        var b = 0;
        var c = new List$1(Rectangle.$, 0);
        try {
            for (var _c = __values(fromEnum(this.c8.a)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var d = _d.value;
                c.add(d);
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        c.aa(function (e, f) {
            if (e.n < f.n) {
                return -1;
            }
            else if (e.n > f.n) {
                return 1;
            }
            else {
                return 0;
            }
        });
        try {
            for (var _e = __values(fromEnum(c)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var e = _f.value;
                this.e.bc.exportRectangleData(a, e, "column" + b, ["Main", "Fill"]);
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
        b++;
    };
    RangeColumnSeriesView.$t = markType(RangeColumnSeriesView, 'RangeColumnSeriesView', RangeCategorySeriesView.$);
    return RangeColumnSeriesView;
}(RangeCategorySeriesView));
export { RangeColumnSeriesView };
