// igniteui-react-charts/esm5/lib/RateOfChangeAndMomentumIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RateOfChangeAndMomentumIndicatorStrategy } from "./RateOfChangeAndMomentumIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var RateOfChangeAndMomentumIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RateOfChangeAndMomentumIndicator, _super);
    function RateOfChangeAndMomentumIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RateOfChangeAndMomentumIndicator.prototype.bj = function () {
        return new RateOfChangeAndMomentumIndicator();
    };
    RateOfChangeAndMomentumIndicator.prototype.get_abv = function () {
        return new RateOfChangeAndMomentumIndicatorStrategy();
    };
    Object.defineProperty(RateOfChangeAndMomentumIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    RateOfChangeAndMomentumIndicator.prototype.get_ab8 = function () {
        return RateOfChangeAndMomentumIndicator.$;
    };
    Object.defineProperty(RateOfChangeAndMomentumIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RateOfChangeAndMomentumIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(RateOfChangeAndMomentumIndicator.ach));
        },
        set: function (a) {
            this.h(RateOfChangeAndMomentumIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    RateOfChangeAndMomentumIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    RateOfChangeAndMomentumIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    RateOfChangeAndMomentumIndicator.$t = markType(RateOfChangeAndMomentumIndicator, 'RateOfChangeAndMomentumIndicator', StrategyBasedIndicator.$);
    RateOfChangeAndMomentumIndicator.ach = StrategyBasedIndicator.acd(14, RateOfChangeAndMomentumIndicator.$);
    return RateOfChangeAndMomentumIndicator;
}(StrategyBasedIndicator));
export { RateOfChangeAndMomentumIndicator };
