// igniteui-react-charts/esm5/lib/RelativeStrengthIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RelativeStrengthIndexIndicatorStrategy } from "./RelativeStrengthIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var RelativeStrengthIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RelativeStrengthIndexIndicator, _super);
    function RelativeStrengthIndexIndicator() {
        return _super.call(this) || this;
    }
    RelativeStrengthIndexIndicator.prototype.bj = function () {
        return new RelativeStrengthIndexIndicator();
    };
    RelativeStrengthIndexIndicator.prototype.get_aa2 = function () {
        return 1;
    };
    RelativeStrengthIndexIndicator.prototype.get_abv = function () {
        return new RelativeStrengthIndexIndicatorStrategy();
    };
    Object.defineProperty(RelativeStrengthIndexIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    RelativeStrengthIndexIndicator.prototype.get_ab8 = function () {
        return RelativeStrengthIndexIndicator.$;
    };
    Object.defineProperty(RelativeStrengthIndexIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RelativeStrengthIndexIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(RelativeStrengthIndexIndicator.ach));
        },
        set: function (a) {
            this.h(RelativeStrengthIndexIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    RelativeStrengthIndexIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    RelativeStrengthIndexIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    RelativeStrengthIndexIndicator.$t = markType(RelativeStrengthIndexIndicator, 'RelativeStrengthIndexIndicator', StrategyBasedIndicator.$);
    RelativeStrengthIndexIndicator.ach = StrategyBasedIndicator.acd(14, RelativeStrengthIndexIndicator.$);
    return RelativeStrengthIndexIndicator;
}(StrategyBasedIndicator));
export { RelativeStrengthIndexIndicator };
