// igniteui-react-charts/esm5/lib/RingSeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { RingSeriesBase } from "./RingSeriesBase";
import { Ring } from "./Ring";
import { ArcItem } from "./ArcItem";
import { fromEnum, markType } from "igniteui-react-core";
import { RingCollection } from "./RingCollection";
/**
 * @hidden
 */
var RingSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RingSeries, _super);
    function RingSeries() {
        var _this = _super.call(this) || this;
        _this.eo = null;
        _this.eo = ((function () {
            var $ret = new Ring();
            $ret.c = _this;
            $ret.e = false;
            return $ret;
        })());
        var a = ((function () {
            var $ret = new ArcItem();
            $ret.k = _this.ba;
            $ret.d = _this.eo;
            $ret.n = _this.cs;
            $ret.c = _this.ap;
            $ret.j = _this.a8;
            return $ret;
        })());
        _this.eo.g.add(a);
        _this.ab = RingSeries.$;
        return _this;
    }
    RingSeries.prototype.c4 = function (a, b, c, d) {
        _super.prototype.c4.call(this, a, b, c, d);
        if (b == RingSeriesBase.$$p[6]) {
            for (var e = 0; e < this.eo.ringControl.arcs.count; e++) {
                this.eo.ringControl.arcs.item(e).ef = d;
            }
        }
        if (b == RingSeriesBase.$$p[7]) {
            for (var f = 0; f < this.eo.ringControl.arcs.count; f++) {
                this.eo.ringControl.arcs.item(f).ag = d;
            }
        }
        if (b == RingSeriesBase.$$p[28]) {
            for (var g = 0; g < this.eo.ringControl.arcs.count; g++) {
                this.eo.ringControl.arcs.item(g).fa = d;
            }
        }
        if (b == RingSeriesBase.$$p[29]) {
            for (var h = 0; h < this.eo.ringControl.arcs.count; h++) {
                this.eo.ringControl.arcs.item(h).aj = d;
            }
        }
        if (b == RingSeriesBase.$$p[18]) {
            for (var i = 0; i < this.eo.ringControl.arcs.count; i++) {
                this.eo.ringControl.arcs.item(i).ew = d;
            }
        }
        if (b == RingSeriesBase.$$p[19]) {
            for (var j = 0; j < this.eo.ringControl.arcs.count; j++) {
                this.eo.ringControl.arcs.item(j).ah = d;
            }
        }
        if (b == RingSeriesBase.$$p[21]) {
            for (var k = 0; k < this.eo.ringControl.arcs.count; k++) {
                this.eo.ringControl.arcs.item(k).e1 = d;
            }
        }
        if (b == RingSeriesBase.$$p[22]) {
            for (var l = 0; l < this.eo.ringControl.arcs.count; l++) {
                this.eo.ringControl.arcs.item(l).ai = d;
            }
        }
        if (b == RingSeriesBase.$$p[2]) {
            for (var m = 0; m < this.eo.ringControl.arcs.count; m++) {
                this.eo.ringControl.arcs.item(m).a5 = d;
            }
        }
        if (b == RingSeriesBase.$$p[3]) {
            for (var n = 0; n < this.eo.ringControl.arcs.count; n++) {
                this.eo.ringControl.arcs.item(n).a6 = d;
            }
        }
        if (b == RingSeriesBase.$$p[20]) {
            for (var o = 0; o < this.eo.ringControl.arcs.count; o++) {
                this.eo.ringControl.arcs.item(o).ez = this.b3;
            }
        }
        if (b == RingSeriesBase.$$p[0]) {
            this.c2();
        }
        if (b == RingSeriesBase.$$p[33]) {
            if (this.eo.g != null && this.eo.g.count > 0) {
                this.eo.g._inner[0].k = this.ba;
            }
        }
        if (b == RingSeriesBase.$$p[36]) {
            if (this.eo.g != null && this.eo.g.count > 0) {
                this.eo.g._inner[0].n = this.cs;
                this.c3();
            }
        }
        if (this.eo != null && this.aw.e()) {
            this.eo.o();
            if (this.eo.f()) {
                this.eo.c.aw.l();
            }
        }
    };
    RingSeries.prototype.au = function () {
        var a = new RingCollection();
        if (this.eo.g._inner[0].g.count > 0) {
            a.add(this.eo);
        }
        return a;
    };
    RingSeries.prototype.c3 = function () {
        if (this.eo != null) {
            this.eo.g._inner[0].h = this.itemsSource;
            this.eo.g._inner[0].p(this.ba);
            if (this.ax != null) {
                this.ax.bx();
                this.ax.b2();
            }
        }
    };
    RingSeries.prototype.c2 = function () {
        if (this.eo != null) {
            this.eo.g._inner[0].a = this.ak;
        }
    };
    RingSeries.prototype.c8 = function () {
        if (this.eo != null) {
            this.width = this.eo.q.width;
            this.height = this.eo.q.height;
            this.aw.m(this.eo.p.x, this.eo.p.y);
        }
    };
    RingSeries.prototype.c7 = function () {
        var e_1, _a;
        if (this.eo != null) {
            try {
                for (var _b = __values(fromEnum(this.eo.ringControl.arcs.a)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var a = _c.value;
                    a.gy();
                }
            }
            catch (e_1_1) {
                e_1 = { error: e_1_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_1)
                        throw e_1.error;
                }
            }
        }
    };
    RingSeries.prototype.cy = function (a, b) {
        _super.prototype.cy.call(this, a, b);
        if (this.eo != null) {
            this.eo.l(this.itemsSource, a, b);
        }
    };
    RingSeries.prototype.cx = function () {
        _super.prototype.cx.call(this);
        if (this.eo != null) {
            this.eo.k(this.itemsSource);
        }
    };
    RingSeries.prototype.cz = function (a, b) {
        _super.prototype.cz.call(this, a, b);
        if (this.eo != null) {
            this.eo.m(this.itemsSource, a, b);
        }
    };
    RingSeries.prototype.c0 = function (a, b, c) {
        _super.prototype.c0.call(this, a, b, c);
        if (this.eo != null) {
            this.eo.n(this.itemsSource, a, b, c);
        }
    };
    RingSeries.$t = markType(RingSeries, 'RingSeries', RingSeriesBase.$);
    return RingSeries;
}(RingSeriesBase));
export { RingSeries };
