// igniteui-react-charts/esm5/lib/ScatterPolylineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ShapeSeriesBase } from "./ShapeSeriesBase";
import { StyleSelector } from "./StyleSelector";
import { DependencyProperty } from "igniteui-react-core";
import { Style } from "igniteui-react-core";
import { markType, markDep, typeCast } from "igniteui-react-core";
import { ShapeSeriesViewBase } from "./ShapeSeriesViewBase";
import { ChartSelection } from "./ChartSelection";
import { List$1 } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { Path } from "igniteui-react-core";
import { stringIsNullOrEmpty } from "igniteui-react-core";
/**
 * @hidden
 */
var ScatterPolylineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolylineSeries, _super);
    function ScatterPolylineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = ScatterPolylineSeries.$;
        return _this;
    }
    ScatterPolylineSeries.prototype.bj = function () {
        return new ScatterPolylineSeries();
    };
    ScatterPolylineSeries.prototype.get_fr = function () {
        return true;
    };
    ScatterPolylineSeries.prototype.cs = function () {
        return new ScatterPolylineSeriesView(this);
    };
    ScatterPolylineSeries.prototype.get_x0 = function () {
        return false;
    };
    ScatterPolylineSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.z7 = a;
    };
    ScatterPolylineSeries.prototype.get_x6 = function () {
        return false;
    };
    ScatterPolylineSeries.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case ScatterPolylineSeries.$$p[0]:
                this.z7.cy(this.aad);
                this.rr(false);
                this.ql();
                break;
            case ScatterPolylineSeries.$$p[1]:
                this.z7.cz(this.z8);
                this.rr(false);
                this.ql();
                break;
        }
    };
    ScatterPolylineSeries.prototype.zn = function (a) {
        var b = a;
        b.cy(this.aad);
        b.cz(this.z8);
        _super.prototype.zn.call(this, a);
    };
    ScatterPolylineSeries.prototype.a2 = function (a, b) {
        var c = this.z7.c3(a, b, true);
        if (c != null) {
            return c.dataContext;
        }
        return _super.prototype.a2.call(this, a, b);
    };
    ScatterPolylineSeries.prototype.ak = function (a, b) {
        var c = new List$1(ChartSelection.$, 0);
        this.p4(a, c, b);
        return c.toArray();
    };
    ScatterPolylineSeries.prototype.p4 = function (a, b, c) {
        this.zf(a, b, c);
    };
    ScatterPolylineSeries.$t = markType(ScatterPolylineSeries, 'ScatterPolylineSeries', ShapeSeriesBase.$);
    ScatterPolylineSeries.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterPolylineSeries, 'raisePropertyChanged', ['ShapeStyle:aad:aab', [Style.$, null], 'ShapeStyleSelector:z8:aac', [StyleSelector.$, null]]);
    return ScatterPolylineSeries;
}(ShapeSeriesBase));
export { ScatterPolylineSeries };
/**
 * @hidden
 */
var ScatterPolylineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolylineSeriesView, _super);
    function ScatterPolylineSeriesView(a) {
        return _super.call(this, a) || this;
    }
    ScatterPolylineSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.b6.jp = Defaults.aa;
        }
    };
    ScatterPolylineSeriesView.prototype.c5 = function (a, b) {
        return this.c6(a, b, false);
    };
    ScatterPolylineSeriesView.prototype.cm = function (a, b, c) {
        _super.prototype.cm.call(this, a, b, c);
        ShapeSeriesViewBase.cl(this, a, b);
    };
    ScatterPolylineSeriesView.prototype.ck = function (a, b) {
        var c = typeCast(Path.$, a);
        if (c == null) {
            return;
        }
        c.an = b;
    };
    ScatterPolylineSeriesView.prototype.ac = function (a) {
        var b = _super.prototype.ac.call(this, a);
        b._fill = null;
        return b;
    };
    ScatterPolylineSeriesView.prototype.ab = function (a) {
        var b = _super.prototype.ab.call(this, a);
        b._fill = null;
        return b;
    };
    ScatterPolylineSeriesView.prototype.cb = function (a) {
        return a.width >= this.b6.yn || a.height >= this.b6.yn;
    };
    ScatterPolylineSeriesView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span";
        var b = this.e.ba.a$u(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.b6.title + "</span><br/>";
        a += "</div>";
        return a;
    };
    ScatterPolylineSeriesView.$t = markType(ScatterPolylineSeriesView, 'ScatterPolylineSeriesView', ShapeSeriesViewBase.$);
    return ScatterPolylineSeriesView;
}(ShapeSeriesViewBase));
export { ScatterPolylineSeriesView };
