// igniteui-react-charts/esm5/lib/ScatterBase_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { MarkerSeries } from "./MarkerSeries";
import { ISupportsErrorBars_$type } from "./ISupportsErrorBars";
import { ITrendLineSeries_$type } from "./ITrendLineSeries";
import { ScatterAxisInfoCache } from "./ScatterAxisInfoCache";
import { NumericXAxis } from "./NumericXAxis";
import { DependencyProperty } from "igniteui-react-core";
import { NumericYAxis } from "./NumericYAxis";
import { Series } from "./Series";
import { TrendLineType_$type } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
import { PenLineCap_$type } from "igniteui-react-core";
import { DoubleCollection } from "igniteui-react-core";
import { CollisionAvoidanceType_$type } from "./CollisionAvoidanceType";
import { ScatterItemSearchMode_$type } from "./ScatterItemSearchMode";
import { ScatterErrorBarSettings } from "./ScatterErrorBarSettings";
import { HighlightingInfo } from "igniteui-react-core";
import { ScatterFrame } from "./ScatterFrame";
import { Base, runOn, delegateCombine, enumGetBox, Number_$type, typeCast, fromEnum, Point_$type, markType, markDep, PointUtil } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { ScalerParams } from "./ScalerParams";
import { Rect } from "igniteui-react-core";
import { Marker } from "./Marker";
import { MarkerSeriesView } from "./MarkerSeriesView";
import { SeriesLayerPropertyOverlay } from "./SeriesLayerPropertyOverlay";
import { IFastItemsSource_$type } from "igniteui-react-core";
import { Clipper } from "igniteui-react-core";
import { OwnedPoint } from "./OwnedPoint";
import { Dictionary$2 } from "igniteui-react-core";
import { AxisRange } from "./AxisRange";
import { GeometryUtil } from "igniteui-react-core";
import { HashPool$2 } from "igniteui-react-core";
import { DataContext } from "igniteui-react-core";
import { TrendResolutionParams } from "./TrendResolutionParams";
import { ScatterTrendLineManager } from "./ScatterTrendLineManager";
import { Thickness } from "igniteui-react-core";
import { Size } from "igniteui-react-core";
import { AssigningScatterStyleEventArgs } from "./AssigningScatterStyleEventArgs";
import { AssigningScatterMarkerStyleEventArgs } from "./AssigningScatterMarkerStyleEventArgs";
import { ChartSelection } from "./ChartSelection";
import { SeriesRenderManager } from "./SeriesRenderManager";
import { ErrorBarsHelper } from "./ErrorBarsHelper";
import { PathGeometry } from "igniteui-react-core";
import { DataLegendSeriesValueInfo } from "igniteui-react-core";
import { DataLegendSeriesContext } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { truncate, isNaN_, isInfinity } from "igniteui-react-core";
import { boxArray$1, arrayShallowClone } from "igniteui-react-core";
import { Path } from "igniteui-react-core";
import { DataTemplateMeasureInfo } from "igniteui-react-core";
import { NumericMarkerManager } from "./NumericMarkerManager";
import { stringIsNullOrEmpty } from "igniteui-react-core";
/**
 * @hidden
 */
var ScatterBase = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterBase, _super);
    function ScatterBase() {
        var _this = _super.call(this) || this;
        _this.yw = null;
        _this.zf = null;
        _this.zg = null;
        _this.acl = null;
        _this.acm = null;
        _this.y5 = 1;
        _this.ym = null;
        _this.yn = null;
        _this.y2 = null;
        _this.y4 = null;
        _this.y0 = null;
        _this.zl = false;
        _this.zk = false;
        _this.zd = null;
        _this.ze = null;
        _this.assigningScatterStyle = null;
        _this.assigningScatterMarkerStyle = null;
        _this.y9 = null;
        _this.y3 = null;
        _this.yz = null;
        _this.zj = true;
        _this.zy = 0;
        _this.zz = 0;
        _this.z0 = 0;
        _this.z1 = 0;
        _this.y6 = _this.y7();
        _this.y9 = new ScatterSeriesRenderManager();
        _this.y3 = new ScatterFrame();
        _this.acm = Rect.empty;
        _this.acl = Rect.empty;
        return _this;
    }
    ScatterBase.prototype.cs = function () {
        return new ScatterBaseView(this);
    };
    ScatterBase.prototype.o2 = function () {
        _super.prototype.o2.call(this);
        this.xAxis = null;
        this.yAxis = null;
    };
    ScatterBase.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.yx = a;
    };
    ScatterBase.prototype.get_fw = function () {
        return true;
    };
    ScatterBase.prototype.ai = function () {
        return [this.xAxis, this.yAxis];
    };
    ScatterBase.prototype.kq = function (a, b) {
        var c = this.nd(b);
        if (c == ScatterBase.$$p[12]) {
            return this.kr(a, b, this.aat);
        }
        if (c == ScatterBase.$$p[16]) {
            return this.kr(a, b, this.aa1);
        }
        if (c == ScatterBase.$$p[1]) {
            return this.kr(a, b, this.aad);
        }
        if (c == ScatterBase.$$p[2]) {
            return this.kr(a, b, this.aaf);
        }
        return null;
    };
    ScatterBase.prototype.md = function (a) {
        var b = this.nd(a);
        if (b == ScatterBase.$$p[12]) {
            return this.aat;
        }
        if (b == ScatterBase.$$p[16]) {
            return this.aa1;
        }
        if (b == ScatterBase.$$p[1]) {
            return this.aad;
        }
        if (b == ScatterBase.$$p[2]) {
            return this.aaf;
        }
        return _super.prototype.md.call(this, a);
    };
    ScatterBase.prototype.eb = function (a, b, c, d, e) {
        var _this = this;
        if (!this.hasMarkers) {
            return false;
        }
        var f = this.xAxis;
        var g = new ScalerParams(1, this.db.v4, this.ws, f.ch);
        g.c = this.getEffectiveViewport1(this.cv);
        var h = this.yAxis;
        var i = new ScalerParams(1, this.db.v4, this.ws, h.ch);
        i.c = this.getEffectiveViewport1(this.cv);
        a.clear();
        this.w6.cj(function (j) {
            if (j._visibility == 0) {
                var k = new SeriesPointOfInterest();
                var l = j.n;
                var m = j.o;
                if (l < _this.ws.left || l > _this.ws.right || m < _this.ws.top || m > _this.ws.bottom) {
                    return;
                }
                k.d = f.eo(l, g);
                k.e = h.eo(m, i);
                k.b = j.content;
                k.g = _this.vg;
                k.c = j.ak;
                if (_this.g9) {
                    k.g = _this.vn(j.ak);
                }
                a.add(k);
            }
        });
        return true;
    };
    ScatterBase.prototype.b2 = function () {
        if (this.b3 == 0) {
            if (this.z9 <= 2000) {
                return 1;
            }
            else {
                return _super.prototype.b2.call(this);
            }
        }
        else {
            return _super.prototype.b2.call(this);
        }
    };
    ScatterBase.prototype.wp = function (a) {
        if (!this.hasMarkers || !this.xg()) {
            return Rect.empty;
        }
        var b = this.w6.b1(this.v7(a));
        if (b != null) {
            return this.w6.cu(b);
        }
        return Rect.empty;
    };
    ScatterBase.prototype.gw = function (a, b) {
        if (this.xh(a, b)) {
            return true;
        }
        return false;
    };
    Object.defineProperty(ScatterBase.prototype, "zh", {
        get: function () {
            return this.zf;
        },
        set: function (a) {
            if (this.zf != a) {
                var b = this.zh;
                this.zf = a;
                this.raisePropertyChanged("XColumn", b, this.zh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBase.prototype, "zi", {
        get: function () {
            return this.zg;
        },
        set: function (a) {
            if (this.zg != a) {
                var b = this.zi;
                this.zg = a;
                this.raisePropertyChanged("YColumn", b, this.zi);
            }
        },
        enumerable: false,
        configurable: true
    });
    ScatterBase.prototype.gb = function () {
        return this.highlightedItemsSource != null || (this.aad != null && this.aaf != null);
    };
    ScatterBase.prototype.zw = function () {
        return false;
    };
    ScatterBase.prototype.ht = function () {
        var a = _super.prototype.ht.call(this);
        {
            var b = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "YMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedYMemberPath";
                $ret.q = true;
                return $ret;
            })());
            b.valueResolving = delegateCombine(b.valueResolving, runOn(this, this.aa9));
            a.add(b);
        }
        {
            var c = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "XMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedXMemberPath";
                $ret.q = true;
                return $ret;
            })());
            c.valueResolving = delegateCombine(c.valueResolving, runOn(this, this.aa8));
            a.add(c);
        }
        if (this.zw()) {
            var d = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "DashArray";
                $ret.p = true;
                $ret.x = ((function () {
                    var $ret = new DoubleCollection();
                    $ret.add(5);
                    $ret.add(5);
                    return $ret;
                })());
                $ret.q = true;
                return $ret;
            })());
            a.add(d);
        }
        if (!this.zw()) {
            var e = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Brush";
                $ret.p = true;
                $ret.y = "ActualBrush";
                $ret.q = true;
                return $ret;
            })());
            e.valueResolving = delegateCombine(e.valueResolving, runOn(this, this.r0));
            a.add(e);
        }
        if (!this.zw()) {
            var f = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Outline";
                $ret.p = true;
                $ret.y = "ActualOutline";
                $ret.q = true;
                return $ret;
            })());
            f.valueResolving = delegateCombine(f.valueResolving, runOn(this, this.r0));
            a.add(f);
        }
        return a;
    };
    ScatterBase.prototype.aa8 = function (a, b) {
        if (this.aad != null) {
            b.value = this.aad;
            return;
        }
        b.value = this.aat;
    };
    ScatterBase.prototype.aa9 = function (a, b) {
        if (this.aaf != null) {
            b.value = this.aaf;
            return;
        }
        b.value = this.aa1;
    };
    Object.defineProperty(ScatterBase.prototype, "aci", {
        get: function () {
            return this.c(ScatterBase.abt);
        },
        set: function (a) {
            this.h(ScatterBase.abt, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBase.prototype, "GetErrorBarsXAxis", {
        get: function () {
            return this.xAxis;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBase.prototype, "GetErrorBarsYAxis", {
        get: function () {
            return this.yAxis;
        },
        enumerable: false,
        configurable: true
    });
    ScatterBase.prototype.qb = function () {
        _super.prototype.qb.call(this);
        if (this.xAxis != null) {
            this.xAxis.ib(false);
        }
        if (this.yAxis != null) {
            this.yAxis.ib(false);
        }
    };
    ScatterBase.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        if (this.yx.c2.g$b(a, b, c, d)) {
            this.rr(false);
            this.ql();
        }
        switch (b) {
            case ScatterBase.$$p[10]:
            case ScatterBase.$$p[14]:
            case ScatterBase.$$p[11]:
            case ScatterBase.$$p[15]:
                this.qf();
                break;
            case "FastItemsSource":
                if (typeCast(IFastItemsSource_$type, c) != null) {
                    c.deregisterColumn(this.zh);
                    c.deregisterColumn(this.zi);
                    this.zh = null;
                    this.zi = null;
                }
                if (typeCast(IFastItemsSource_$type, d) != null) {
                    this.zh = this.dk(this.aat);
                    this.zi = this.dk(this.aa1);
                }
                if ((this.yAxis != null && !this.yAxis.c1()) || (this.xAxis != null && !this.xAxis.c1())) {
                    this.rr(this.gg && c != null);
                }
                break;
            case ScatterBase.$$p[9]:
                if (c != null) {
                    c.b0(this);
                }
                if (d != null) {
                    d.ct(this);
                }
                if ((this.xAxis != null && !this.xAxis.c1()) || (d == null && c != null)) {
                    this.rr(false);
                }
                this.ql();
                break;
            case ScatterBase.$$p[13]:
                if (c != null) {
                    c.b0(this);
                }
                if (d != null) {
                    d.ct(this);
                }
                if ((this.yAxis != null && !this.yAxis.c1()) || (d == null && c != null)) {
                    this.rr(false);
                }
                this.ql();
                break;
            case ScatterBase.$$p[12]:
                if (this.dn != null) {
                    this.dn.deregisterColumn(this.zh);
                    this.zh = this.dk(this.aat);
                }
                break;
            case "XColumn":
                this.yx.c2.g$i();
                if (this.xAxis != null && !this.xAxis.c1()) {
                    this.rr(this.gg && this.d7);
                }
                break;
            case ScatterBase.$$p[16]:
                if (this.dn != null) {
                    this.dn.deregisterColumn(this.zi);
                    this.zi = this.dk(this.aa1);
                }
                break;
            case "YColumn":
                this.yx.c2.g$i();
                if (this.yAxis != null && !this.yAxis.c1()) {
                    this.rr(this.gg && this.d7);
                }
                break;
            case ScatterBase.$$p[7]:
                this.rr(false);
                this.ql();
                break;
            case ScatterBase.$$p[8]:
                this.rr(false);
                this.ql();
                break;
            case "TransitionProgress":
                this.y4.c(this.jq, this.y2, this.y0);
                this.aa3();
                try {
                    if (this.d5(this.cv)) {
                        return;
                    }
                    if ((Math.round(this.jq * 100000) / 100000) == 1) {
                        this.abq(this.y0, this.yx);
                    }
                    else {
                        this.abq(this.y4, this.yx);
                    }
                    if (this.db != null) {
                        this.db.qo(this);
                    }
                }
                finally {
                    this.abr();
                }
                break;
            case "TrendLineBrush":
                this.rf(b);
                break;
            case ScatterBase.$$p[0]:
                if (this.yy != null) {
                    this.yy.r = this;
                }
                this.rr(false);
                this.ql();
                break;
            case "TrendLineType":
                this.ql();
                break;
            case ScatterBase.$$p[4]:
                this.zl = this.zt();
                this.rr(false);
                break;
            case ScatterBase.$$p[3]:
                this.zk = this.zs();
                this.rr(false);
                break;
            case ScatterBase.$$p[5]:
                this.y6 = this.y7();
                break;
        }
    };
    ScatterBase.prototype.zt = function () {
        return this.zp;
    };
    ScatterBase.prototype.zs = function () {
        return this.zo;
    };
    ScatterBase.prototype.zn = function (a) {
        if (typeCast(NumericYAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    ScatterBase.prototype.zm = function (a) {
        if (typeCast(NumericXAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    ScatterBase.prototype.abr = function () {
        this.acl = Rect.empty;
        this.acm = Rect.empty;
    };
    ScatterBase.prototype.aa3 = function () {
        var a = this.getViewInfo(this.acl, this.acm);
        this.acl = a.p0;
        this.acm = a.p1;
    };
    ScatterBase.prototype.zr = function (a, b) {
        if (b != 3) {
            return true;
        }
        if (a == null) {
            return true;
        }
        if (this.aat == a || this.aa1 == a) {
            return true;
        }
        return false;
    };
    ScatterBase.prototype.pb = function (a, b, c, d) {
        var e = false;
        if (!this.zr(d, a)) {
            return;
        }
        this.yx.c2.g$e(a, b, c, d);
        if (this.xAxis != null && !this.xAxis.c1()) {
            e = true;
        }
        if (this.yAxis != null && !this.yAxis.c1()) {
            e = true;
        }
        if (a == 0 || a == 2 || a == 3 || a == 4) {
            this.zj = true;
        }
        else if (a == 1) {
            for (var f = b; f < b + c; f++) {
                this.zy += this.zh.item(f);
                this.zz += this.zi.item(f);
            }
            this.z0 = this.zy / this.zh.count;
            this.z1 = this.zz / this.zi.count;
        }
        if (e) {
            this.rr(true);
        }
    };
    ScatterBase.prototype.abi = function (a) {
        this.abj(a, null);
    };
    ScatterBase.prototype.abj = function (a, b) {
        var e_1, _a, e_2, _b;
        var c = this.zh != null ? this.zh.count : 0;
        var d = this.zi != null ? this.zi.count : 0;
        var e = Math.min(c, d);
        if (e <= this.z9) {
            a.m.clear();
            var f = new List$1(OwnedPoint.$, 0);
            try {
                for (var _c = __values(fromEnum(a.u.values)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var g = _d.value;
                    f.add(g);
                }
            }
            catch (e_1_1) {
                e_1 = { error: e_1_1 };
            }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return))
                        _a.call(_c);
                }
                finally {
                    if (e_1)
                        throw e_1.error;
                }
            }
            var h_1 = this.dn;
            f.aa(function (i, j) {
                var k = h_1.indexOf(i.a);
                var l = h_1.indexOf(j.a);
                if (k < l) {
                    return -1;
                }
                if (k > l) {
                    return 1;
                }
                return 0;
            });
            if (b != null) {
                b.i = a.m;
            }
            try {
                for (var _e = __values(fromEnum(f)), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var i = _f.value;
                    if (h_1.indexOf(i.a) >= 0) {
                        if (b != null) {
                            b.j(i.c);
                        }
                        else {
                            a.m.add({ $type: Point_$type, x: i.c.x, y: i.c.y });
                        }
                    }
                }
            }
            catch (e_2_1) {
                e_2 = { error: e_2_1 };
            }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return))
                        _b.call(_e);
                }
                finally {
                    if (e_2)
                        throw e_2.error;
                }
            }
        }
    };
    ScatterBase.prototype.ay = function (a) {
        if (a != null && a == this.xAxis && this.zh != null) {
            return new AxisRange(this.zh.minimum, this.zh.maximum);
        }
        if (a != null && a == this.yAxis && this.zi != null) {
            return new AxisRange(this.zi.minimum, this.zi.maximum);
        }
        return null;
    };
    ScatterBase.prototype.kn = function (a) {
        var b = this.j3(a);
        if (this.dn != null && b >= 0 && b < this.dn.count) {
            return this.dn.item(b);
        }
        return null;
    };
    Object.defineProperty(ScatterBase.prototype, "y6", {
        get: function () {
            return this.y5;
        },
        set: function (a) {
            var b = this.y5;
            this.y5 = a;
            if (b != this.y5) {
                this.raisePropertyChanged("ActualItemSearchMode", enumGetBox(ScatterItemSearchMode_$type, b), enumGetBox(ScatterItemSearchMode_$type, this.y5));
            }
        },
        enumerable: false,
        configurable: true
    });
    ScatterBase.prototype.y7 = function () {
        if (this.y8 != 0) {
            return this.y8;
        }
        return this.zq ? 4 : 1;
    };
    ScatterBase.prototype.j3 = function (a) {
        var b = truncate(Math.round(this.ir(a)));
        return b;
    };
    ScatterBase.prototype.ir = function (a) {
        switch (this.y6) {
            case 1: return this.z5(a);
            case 3: return this.z6(a);
            case 2: return this.z3(a);
            case 4: return this.z4(a);
            case 5: return this.z2(a);
        }
        return -1;
    };
    ScatterBase.prototype.z4 = function (a) {
        return this.z7(a, false);
    };
    ScatterBase.prototype.z3 = function (a) {
        var b = this.v7(a);
        var c = b.x;
        var d = b.y;
        if (this.dn == null || this.xAxis == null || this.yAxis == null || this.zh == null || this.zi == null) {
            return -1;
        }
        if (this.dn.count > this.z8) {
            return -1;
        }
        this.aa6();
        var e = this.ym;
        var f = this.yn;
        if (e == null || f == null) {
            return -1;
        }
        var g = 1.7976931348623157E+308;
        var h = -1;
        for (var i = 0; i < this.zh.count; i++) {
            var j = e[i];
            var k = f[i];
            var l = (c - j) * (c - j) + (d - k) * (d - k);
            if (l < g) {
                g = l;
                h = i;
            }
        }
        return h;
    };
    ScatterBase.prototype.aa6 = function () {
        var _this = this;
        var a = this.cv.by;
        var b = this.cv.bx;
        var c = this.getEffectiveViewport();
        if (this.yw == null) {
            return;
        }
        var d = new ScalerParams(0, a, b, this.yw.d, c);
        var e = ((function () {
            var $ret = new ScalerParams(0, a, b, _this.yw.e, c);
            $ret.b = _this.iy();
            return $ret;
        })());
        if (this.ym == null || this.yn == null) {
            var f = arrayShallowClone(this.zh.asArray());
            var g = arrayShallowClone(this.zi.asArray());
            this.xAxis.hq(boxArray$1(f), 0, f.length, d);
            this.yAxis.hq(boxArray$1(g), 0, g.length, e);
            this.ym = f;
            this.yn = g;
        }
    };
    ScatterBase.prototype.get_zq = function () {
        return false;
    };
    Object.defineProperty(ScatterBase.prototype, "zq", {
        get: function () {
            return this.get_zq();
        },
        enumerable: false,
        configurable: true
    });
    ScatterBase.prototype.z7 = function (a, b) {
        if (!this.zq) {
            return -1;
        }
        var c = { $type: Point_$type, x: 0, y: 0 };
        var d = true;
        var e = this.y1();
        if (e == null) {
            return -1;
        }
        if (e.m == null) {
            return -1;
        }
        if (this.dn == null || this.dn.count > this.z8) {
            return -1;
        }
        this.aa6();
        var f = this.ym;
        var g = this.yn;
        if (f == null || g == null) {
            return -1;
        }
        var h = this.v7(a);
        var i = 1.7976931348623157E+308;
        var j = -1;
        for (var k = 0; k < this.ym.length; k++) {
            var l = { $type: Point_$type, x: this.ym[k], y: this.yn[k] };
            if (k > 0) {
                if (!b || e.j.containsKey(this.dn.item(k)) || e.j.containsKey(this.dn.item(k - 1))) {
                    var m = GeometryUtil.h(h, c, l);
                    if (m < i) {
                        i = m;
                        j = k;
                    }
                }
            }
            c = l;
        }
        if (j >= 0) {
            var n = { $type: Point_$type, x: this.ym[j], y: this.yn[j] };
            var o = { $type: Point_$type, x: this.ym[j - 1], y: this.yn[j - 1] };
            if (!b || (e.j.containsKey(this.dn.item(j)) && e.j.containsKey(this.dn.item(j - 1)))) {
                var p = GeometryUtil.m(h, o, n);
                return (j - 1) + p;
            }
            else if (e.j.containsKey(this.dn.item(j))) {
                return j;
            }
            else {
                return j - 1;
            }
        }
        return -1;
    };
    ScatterBase.prototype.z6 = function (a) {
        return this.z7(a, true);
    };
    ScatterBase.prototype.z2 = function (a) {
        var _this = this;
        var b = this.v7(a);
        var c = b.x;
        var d = b.y;
        if (this.yx.c3.e > this.z8) {
            return -1;
        }
        var e = -1;
        this.w6.ck(function (f) {
            var g = _this.yx.cu(f);
            var h = g.left;
            var i = g.top;
            var j = g.width;
            var k = g.height;
            if (_this.xd) {
                var l = h + j / 2;
                var m = i + k / 2;
                var n = (j / 2) * (j / 2);
                var o = (c - l) * (c - l) + (d - m) * (d - m);
                if (o < n) {
                    if (e == -1) {
                        e = f.ak;
                    }
                }
            }
            else {
                if (c > h && c < h + j && d > i && d < i + k) {
                    if (e == -1) {
                        e = f.ak;
                    }
                }
            }
        });
        return e;
    };
    ScatterBase.prototype.z5 = function (a) {
        var _this = this;
        var b = null;
        var c = 1.7976931348623157E+308;
        var d = this.v7(a);
        var e = d.x;
        var f = d.y;
        if (this.yx.c3.e > this.z8) {
            return -1;
        }
        this.yx.da(function (g) {
            if (g._visibility == 0) {
                var h = _this.yx.cu(g);
                var i = h.left + h.width / 2;
                var j = h.top + h.height / 2;
                var k = (e - i) * (e - i) + (f - j) * (f - j);
                if (k < c) {
                    c = k;
                    b = g;
                }
            }
        });
        if (b != null) {
            return b.ak;
        }
        return -1;
    };
    ScatterBase.prototype.wc = function (a, b, c) {
        var _this = this;
        if (this.xAxis == null || this.yAxis == null || this.zh == null || this.zi == null) {
            return { $type: Point_$type, x: NaN, y: NaN };
        }
        var d = this.cv.by;
        var e = this.cv.bx;
        var f = this.getEffectiveViewport();
        var g = new ScalerParams(0, d, e, this.xAxis.ch, f);
        var h = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yAxis.ch, f);
            $ret.b = _this.iy();
            return $ret;
        })());
        if (b) {
            var i = this.ir(a);
            if (i == -1) {
                return { $type: Point_$type, x: NaN, y: NaN };
            }
            var j = truncate(Math.floor(i));
            var k = truncate(Math.ceil(i));
            if (j < 0) {
                j = 0;
            }
            if (k < 0) {
                k = 0;
            }
            if (k > this.dn.count - 1) {
                k = this.dn.count - 1;
            }
            if (j > this.dn.count - 1) {
                j = this.dn.count - 1;
            }
            if (k == j) {
                var l = this.xAxis.el(this.zh.item(j), g);
                var m = this.yAxis.el(this.zi.item(j), h);
                return { $type: Point_$type, x: l, y: m };
            }
            var n = i - Math.floor(i);
            var o = this.xAxis.el(this.zh.item(j), g);
            var p = this.yAxis.el(this.zi.item(j), h);
            var q = this.xAxis.el(this.zh.item(k), g);
            var r = this.yAxis.el(this.zi.item(k), h);
            var s = q - o;
            var t = r - p;
            return { $type: Point_$type, x: o + s * n, y: p + t * n };
        }
        else {
            var u = this.j3(a);
            if (u == -1) {
                return { $type: Point_$type, x: NaN, y: NaN };
            }
            var v = this.xAxis.el(this.zh.item(u), g);
            var w = this.yAxis.el(this.zi.item(u), h);
            return { $type: Point_$type, x: v, y: w };
        }
    };
    ScatterBase.prototype.gd = function (a) {
        var _this = this;
        var b = this.db != null ? this.db.v4 : Rect.empty;
        b = b.copy();
        var c = this.db != null ? this.db.wk : Rect.empty;
        var d = new Rect(0, 0, 0, 1, 1);
        var e = this.wl(this.cv);
        var f = new ScalerParams(0, d, d, this.xAxis.ch, e);
        var g = ((function () {
            var $ret = new ScalerParams(0, d, d, _this.yAxis.ch, e);
            $ret.b = _this.iy();
            return $ret;
        })());
        var h = !b.isEmpty && !c.isEmpty && this.dn != null ? this.dn.indexOf(a) : -1;
        var i = this.xAxis != null && this.zh != null && h < this.zh.count ? this.xAxis.el(this.zh.item(h), f) : NaN;
        var j = this.yAxis != null && this.zi != null && h < this.zi.count ? this.yAxis.el(this.zi.item(h), g) : NaN;
        if (!isNaN_(i)) {
            if (i < b.left + 0.1 * b.width) {
                i = i + 0.4 * b.width;
                b.x = i - 0.5 * b.width;
            }
            if (i > b.right - 0.1 * b.width) {
                i = i - 0.4 * b.width;
                b.x = i - 0.5 * b.width;
            }
        }
        if (!isNaN_(j)) {
            if (j < b.top + 0.1 * b.height) {
                j = j + 0.4 * b.height;
                b.y = j - 0.5 * b.height;
            }
            if (j > b.bottom - 0.1 * b.height) {
                j = j - 0.4 * b.height;
                b.y = j - 0.5 * b.height;
            }
        }
        if (this.c3 != null) {
            this.c3.au(this.db, b, true);
        }
        return h >= 0;
    };
    ScatterBase.prototype.sq = function (a, b) {
        this.rr(false);
    };
    ScatterBase.prototype.sr = function (a, b) {
        this.rr(false);
    };
    ScatterBase.prototype.y1 = function () {
        var a = this.y0;
        if (this.d3()) {
            if (this.dc.g) {
                this.dc.t();
            }
            a = this.y4;
        }
        return a;
    };
    ScatterBase.prototype.aa4 = function (a, b, c, d, e) {
        var _this = this;
        b.g = new Dictionary$2(Base.$, OwnedPoint.$, 1, c);
        var f = this.getEffectiveViewport1(a);
        var g = this.dn;
        var h;
        var i;
        var j = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yw.d, f);
            $ret.c = _this.getEffectiveViewport1(a);
            return $ret;
        })());
        var k = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yw.e, f);
            $ret.c = _this.getEffectiveViewport1(a);
            $ret.b = _this.iy();
            return $ret;
        })());
        var l = false;
        var m = 0;
        if (this.db != null) {
            if (this.db.f4()) {
                var n = this.db.m4(d, j);
                l = n.c;
                m = n.d;
            }
        }
        var _loop_1 = function (o) {
            h = this_1.zh.item(o);
            i = this_1.zi.item(o);
            var p = { $type: Point_$type, x: this_1.yw.a.el(h, j), y: this_1.yw.b.el(i, k) };
            if (l) {
                if (p.x < e.left || p.x > e.right) {
                    if (p.x + m >= e.left && p.x + m <= e.right) {
                        p = { $type: Point_$type, x: p.x + m, y: i };
                    }
                }
            }
            if (!isInfinity(p.x) && !isInfinity(p.y)) {
                var q_1 = g.item(o);
                if (!b.g.containsKey(q_1)) {
                    var r_1 = { $type: Point_$type, x: h, y: i };
                    var s_1 = { $type: Point_$type, x: p.x, y: p.y };
                    b.g.addItem(q_1, ((function () {
                        var $ret = new OwnedPoint();
                        $ret.a = q_1;
                        $ret.b = r_1;
                        $ret.c = s_1;
                        return $ret;
                    })()));
                }
            }
        };
        var this_1 = this;
        for (var o = 0; o < c; o++) {
            _loop_1(o);
        }
    };
    ScatterBase.prototype.abf = function (a, b) {
        var _this = this;
        this.ym = null;
        this.yn = null;
        a.j.clear();
        a.n.clear();
        a.h.clear();
        a.l.clear();
        a.i.clear();
        a.k.clear();
        var c = Math.min(this.zh != null ? this.zh.count : 0, this.zi != null ? this.zi.count : 0);
        var d = b.by;
        var e = b.bx;
        var f = this.getEffectiveViewport1(b);
        var g = new ScalerParams(0, d, e, this.xAxis.ch, f);
        g.c = this.getEffectiveViewport1(b);
        var h = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yAxis.ch, f);
            $ret.b = _this.iy();
            return $ret;
        })());
        h.c = this.getEffectiveViewport1(b);
        if (c < 1) {
            return;
        }
        this.yw = ((function () {
            var $ret = new ScatterAxisInfoCache();
            $ret.a = _this.xAxis;
            $ret.b = _this.yAxis;
            $ret.d = _this.xAxis.ch;
            $ret.e = _this.yAxis.ch;
            $ret.c = _this.dn;
            return $ret;
        })());
        var i = b;
        if (c <= this.z9) {
            this.aa4(i, a, c, d, e);
        }
        if (this.xg()) {
            b.c0.l(a.j, this.z9, d, e, this.id);
        }
        var j = ((function () {
            var $ret = new Clipper(0, e, false);
            $ret.i = a.n;
            return $ret;
        })());
        var k = this.wj(b);
        var l = this.xAxis.eo(k.left, g);
        var m = this.xAxis.eo(k.right, g);
        b.c2.h(a.n, this.trendLineType, this.zh, this.zi, this.trendLinePeriod, function (n) { return _this.xAxis.el(n, g); }, function (n) { return _this.yAxis.el(n, h); }, ((function () {
            var $ret = new TrendResolutionParams();
            $ret.d = _this.id;
            $ret.i = e;
            $ret.h = f;
            $ret.j = d;
            return $ret;
        })()), j, l, m);
        this.abd(a, b);
    };
    ScatterBase.prototype.o5 = function (a, b) {
        _super.prototype.o5.call(this, a, b);
        var c = b;
        c.c9(a);
    };
    ScatterBase.prototype.get_fe = function () {
        return true;
    };
    ScatterBase.prototype.zb = function (a, b, c) {
        return null;
    };
    ScatterBase.prototype.zc = function (a, b, c) {
        if (this.dn == null) {
            return null;
        }
        var d = this.y1();
        var e = this.yx.b1(this.v7(b));
        var f = -1;
        if (e == null) {
            if (a != null) {
                f = this.dn.indexOf(a);
            }
            else {
                return null;
            }
        }
        else {
            f = e.ak;
        }
        if (f == -1) {
            return null;
        }
        var g = new HighlightingInfo();
        g.i = this;
        g.g = f;
        g.f = f;
        g.c = false;
        g.d = true;
        if (c != null && c.g == g.g && c.f == g.f) {
            return c;
        }
        return g;
    };
    ScatterBase.prototype.wv = function (a, b) {
        if (this.w8 != 1 && this.w8 != 13) {
            var c = this.w6.cv();
            var d = (c.width / 2) + (this.xj / 2);
            var e = (c.height / 2) + (this.xj / 2);
            var f = d;
            var g = d;
            var h = e;
            var i = e;
            if (a == 2 || a == 4) {
                f = 0;
            }
            if (a == 2 || a == 3) {
                g = 0;
            }
            if (b == 2 || b == 3) {
                h = 0;
            }
            if (b == 2 || b == 4) {
                i = 0;
            }
            return new Thickness(1, f, h, g, i);
        }
        return new Thickness(1, 0, 0, 0, 0);
    };
    ScatterBase.prototype.za = function (a, b, c) {
        var d = new HighlightingInfo();
        d.i = this;
        d.g = 0;
        d.f = this.dn.count - 1;
        d.c = true;
        if (c != null && c.g == d.g && c.f == d.f) {
            return c;
        }
        return d;
    };
    ScatterBase.prototype.dg = function (a, b, c) {
        if (this.ed && this.fz(this.bx) && !c) {
            var d = this.zb(a, b, this.zd);
            this.zd = d;
            return d;
        }
        else {
            var e = this.za(a, b, this.zd);
            this.zd = e;
            return e;
        }
    };
    ScatterBase.prototype.dh = function (a, b, c) {
        if (this.fz(this.bx) && !c) {
            var d = this.zc(a, b, this.ze);
            this.ze = d;
            return d;
        }
        else {
            var e = this.za(a, b, this.ze);
            e.d = true;
            this.ze = e;
            return e;
        }
    };
    ScatterBase.prototype.zv = function () {
        return (this.assigningScatterStyle != null && this.zl) || this.d2 || (this.bx != 1) || (this.cp != 1) || (this.co != 1);
    };
    ScatterBase.prototype.zu = function () {
        return (this.assigningScatterMarkerStyle != null && this.zk) || this.d2 || (this.bx != 1) || (this.cp != 1) || (this.co != 1);
    };
    ScatterBase.prototype.abq = function (a, b) {
        var _this = this;
        this.qx();
        var c = b.bx;
        this.yw = ((function () {
            var $ret = new ScatterAxisInfoCache();
            $ret.a = _this.xAxis;
            $ret.b = _this.yAxis;
            $ret.d = _this.xAxis.ch;
            $ret.e = _this.yAxis.ch;
            return $ret;
        })());
        if (this.xg()) {
            b.c0.j(a.j, this.xi);
            if (this.zu()) {
                this.aa5(b, a);
            }
        }
        b.cr();
        var d = ((function () {
            var $ret = new Clipper(1, NaN, c.bottom, NaN, c.top, false);
            $ret.i = b.c2.g$j.an;
            return $ret;
        })());
        b.c2.g$h(a.n, d);
        this.abn(a, b);
    };
    ScatterBase.prototype.ak = function (a, b) {
        var c = new List$1(ChartSelection.$, 0);
        this.p4(a, c, b);
        return c.toArray();
    };
    ScatterBase.prototype.p4 = function (a, b, c) {
        this.aa7(a, b, c, 0, null);
    };
    ScatterBase.prototype.aa7 = function (a, b, c, d, e) {
        var _this = this;
        var f = this.wr(a);
        var g = this.dn.count;
        var h = this.yw.a;
        var i = this.yw.b;
        var j = this.wj(this.cv);
        var k = new ScalerParams(0, f, this.db.wk, h.ch, this.getEffectiveViewport());
        var l = new ScalerParams(0, f, this.db.wk, i.ch, this.getEffectiveViewport());
        if (d == 0) {
            var _loop_2 = function (m) {
                var n = h.el(this_2.zh.item(m), k);
                var o = i.el(this_2.zi.item(m), l);
                if (n < j.width && n > 0 && o < j.height && o > 0) {
                    if (c == 7 || c == 6) {
                        b.add(((function () {
                            var $ret = new ChartSelection();
                            $ret.series = _this;
                            return $ret;
                        })()));
                        return { value: void 0 };
                    }
                    else if (c == 2 || c == 1) {
                        b.add(((function () {
                            var $ret = new ChartSelection();
                            $ret.item = _this.dn.item(m);
                            return $ret;
                        })()));
                    }
                    else {
                        b.add(((function () {
                            var $ret = new ChartSelection();
                            $ret.item = _this.dn.item(m);
                            $ret.series = _this;
                            return $ret;
                        })()));
                    }
                }
            };
            var this_2 = this;
            for (var m = 0; m < g; m++) {
                var state_1 = _loop_2(m);
                if (typeof state_1 === "object")
                    return state_1.value;
            }
        }
        else {
            var _loop_3 = function (p) {
                var q = h.el(this_3.zh.item(p), k);
                var r = i.el(this_3.zi.item(p), l);
                var s = e.item(p) / f.height;
                var t = e.item(p) / f.width;
                var u = new Rect(0, q - t / 2, r - s / 2, t, s);
                if (j.intersectsWith(u)) {
                    if (c == 7 || c == 6) {
                        b.add(((function () {
                            var $ret = new ChartSelection();
                            $ret.series = _this;
                            return $ret;
                        })()));
                        return { value: void 0 };
                    }
                    else if (c == 2 || c == 1) {
                        b.add(((function () {
                            var $ret = new ChartSelection();
                            $ret.item = _this.dn.item(p);
                            return $ret;
                        })()));
                    }
                    else {
                        b.add(((function () {
                            var $ret = new ChartSelection();
                            $ret.item = _this.dn.item(p);
                            $ret.series = _this;
                            return $ret;
                        })()));
                    }
                }
            };
            var this_3 = this;
            for (var p = 0; p < g; p++) {
                var state_2 = _loop_3(p);
                if (typeof state_2 === "object")
                    return state_2.value;
            }
        }
    };
    ScatterBase.prototype.yp = function (a, b) {
        return this.yq(a, b);
    };
    ScatterBase.prototype.yq = function (a, b) {
        if (b < a || a < 0 || a > this.zh.count || b < 0 || b > this.zh.count) {
            return null;
        }
        var c = new Array((b - a) + 1);
        for (var d = a; d <= b; d++) {
            var e = d;
            c[d - a] = this.dn.item(d);
        }
        return c;
    };
    ScatterBase.prototype.aa5 = function (a, b) {
        var _this = this;
        a.b7 = true;
        this.y9.ay(this, this.zu(), runOn(this, this.yp));
        var c = this.y9.d;
        var d = c != null;
        var e = this.getEffectiveViewport1(a);
        var f = this.dn == null ? 0 : this.dn.count;
        this.yx.cj(function (g) {
            if (g._visibility == 0) {
                var h = g.ak;
                var i = g.content;
                if (i == null) {
                    i = new DataContext();
                    g.content = i;
                }
                if (d) {
                    _this.aba(h, f, a.q);
                }
                _this.y9.ac(g, i);
            }
        });
    };
    ScatterBase.prototype.abb = function (a, b, c) {
        var d = this.bx != 1;
        var e = this.cp != 1;
        var f = this.co != 1;
        this.y9.a0(this.y9.b, a, b, false);
        this.y9.ab(this.ib);
        var g = this.y9.ax;
        var h = null;
        var i = null;
        var j = null;
        var k = this.dn;
        if (d && this.db != null) {
            h = this.db.dl.getHighlightingInfo(this, k, g.z, g.y, false);
            g.c = h;
        }
        if (e && this.db != null) {
            h = this.db.dm.getHighlightingInfo(this, k, g.z, g.y, false);
            g.d = h;
        }
        if (f && this.db != null) {
            h = this.db.dk.getHighlightingInfo(this, k, g.z, g.y, false);
            g.b = h;
        }
        g.h = c;
        this.abl(g);
        this.y9.z(this, h, i, j, c, d, e, f, a < 0);
    };
    ScatterBase.prototype.aba = function (a, b, c) {
        var d = this.bx != 1;
        var e = this.cp != 1;
        var f = this.co != 1;
        this.y9.a0(this.y9.d, a, b, true);
        this.y9.aa(this.ib);
        var g = this.y9.aw;
        var h = null;
        var i = null;
        var j = null;
        var k = this.dn;
        var l = null;
        var m = null;
        var n = null;
        if (d && this.db != null) {
            h = this.db.dl.getHighlightingInfo(this, k, g.z, g.y, true);
            if (this.db.dl.isCrossContextHighlight(this)) {
                l = this.db.dl.getHighlightingInfo(this, k, g.z, g.y, true);
            }
            g.c = h;
        }
        if (e && this.db != null) {
            i = this.db.dm.getHighlightingInfo(this, k, g.z, g.y, true);
            if (i == null && this.bn != null) {
                i = this.db.dm.getHighlightingInfo(this.bn, this.bn.dn, g.z, g.y, true);
            }
            if (i == null && this.bk != null) {
                i = this.db.dm.getHighlightingInfo(this.bk, this.bk.dn, g.z, g.y, true);
            }
            if (this.db.dm.isCrossContextHighlight(this)) {
                m = this.db.dm.getHighlightingInfo(this, k, g.z, g.y, true);
            }
            g.d = i;
        }
        if (f && this.db != null) {
            j = this.db.dk.getHighlightingInfo(this, k, g.z, g.y, true);
            if (this.db.dk.isCrossContextHighlight(this)) {
                n = this.db.dk.getHighlightingInfo(this, k, g.z, g.y, true);
            }
            g.d = i;
        }
        g.h = c;
        this.abk(g);
        this.y9.y(this, h, i, j, c, d, e, f, g.v, this.db.dl.totalHighWaterMark, g.x, this.db.dm.totalHighWaterMark, g.t, this.db.dk.totalHighWaterMark);
    };
    ScatterBase.prototype.abl = function (a) {
        if (this.assigningScatterStyle != null && this.zl) {
            this.assigningScatterStyle(this, a);
        }
    };
    ScatterBase.prototype.abk = function (a) {
        if (this.assigningScatterMarkerStyle != null && this.zk) {
            this.assigningScatterMarkerStyle(this, a);
        }
    };
    ScatterBase.prototype.abd = function (a, b) {
        var e_3, _a;
        var c = new ErrorBarsHelper(this, b);
        if (this.yy == null) {
            return;
        }
        var d = this.yy.u;
        var e = this.yy.v;
        var f = 0;
        var g = 0;
        var h = 0;
        var i = 0;
        try {
            for (var _b = __values(fromEnum(a.j.keys)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var j = _c.value;
                var k = a.j.item(j);
                if (d != null) {
                    switch (d.getCalculatorType()) {
                        case 1:
                            var l = this.abg(k, c, d, h, i, true);
                            h = l.p3;
                            i = l.p4;
                            break;
                        case 0:
                            var m = this.abe(k, c, d, h, i, true);
                            h = m.p3;
                            i = m.p4;
                            break;
                        case 2:
                            var n = this.abc(k, c, d, j, h, i, true);
                            h = n.p4;
                            i = n.p5;
                            break;
                        case 3:
                            var o = this.abh(c, d, f, h, i, true);
                            f = o.p2;
                            h = o.p3;
                            i = o.p4;
                            break;
                        case 4:
                            var p = this.abe(k, c, d, h, i, true);
                            h = p.p3;
                            i = p.p4;
                            break;
                    }
                    ;
                    var q = new OwnedPoint();
                    q.c = c.l(d, this.yw.a, k.c, f);
                    q.a = k.a;
                    a.h.addItem(j, q);
                    a.i.addItem(j, [h, i]);
                }
                if (e != null) {
                    switch (e.getCalculatorType()) {
                        case 1:
                            var r = this.abg(k, c, e, h, i, false);
                            h = r.p3;
                            i = r.p4;
                            break;
                        case 0:
                            var s = this.abe(k, c, e, h, i, false);
                            h = s.p3;
                            i = s.p4;
                            break;
                        case 2:
                            var t = this.abc(k, c, e, j, h, i, false);
                            h = t.p4;
                            i = t.p5;
                            break;
                        case 3:
                            var u = this.abh(c, e, g, h, i, false);
                            g = u.p2;
                            h = u.p3;
                            i = u.p4;
                            break;
                        case 4:
                            var v = this.abe(k, c, e, h, i, false);
                            h = v.p3;
                            i = v.p4;
                            break;
                    }
                    ;
                    var w = new OwnedPoint();
                    w.c = c.m(e, this.yw.b, k.c, g);
                    w.a = k.a;
                    a.l.addItem(j, w);
                    a.k.addItem(j, [h, i]);
                }
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
    };
    ScatterBase.prototype.abg = function (a, b, c, d, e, f) {
        if (f) {
            var g = void 0;
            var h = void 0, i = void 0;
            i = this.yw.a;
            if (this.yy.s == 0) {
                g = a.c.x;
                h = this.yw.a;
            }
            else {
                g = a.c.y;
                h = this.yw.b;
            }
            var j = b.f(g, c, h, i, d, e);
            d = j.p4;
            e = j.p5;
        }
        else {
            var k = void 0;
            var l = void 0, m = void 0;
            m = this.yw.b;
            if (this.yy.t == 0) {
                k = a.c.x;
                l = this.yw.a;
            }
            else {
                k = a.c.y;
                l = this.yw.b;
            }
            var n = b.f(k, c, l, m, d, e);
            d = n.p4;
            e = n.p5;
        }
        return {
            p3: d,
            p4: e
        };
    };
    ScatterBase.prototype.abe = function (a, b, c, d, e, f) {
        if (f) {
            var g = b.j(a.c.x, c, this.yw.a, d, e);
            d = g.p3;
            e = g.p4;
        }
        else {
            var h = b.j(a.c.y, c, this.yw.b, d, e);
            d = h.p3;
            e = h.p4;
        }
        return {
            p3: d,
            p4: e
        };
    };
    ScatterBase.prototype.abc = function (a, b, c, d, e, f, g) {
        var h = c.getItemColumn();
        var i = this.dn.indexOf(d);
        if (h != null && i < h.count) {
            var j = h.item(i);
            if (g) {
                var k = b.h(a.c.x, j, this.yw.a, e, f);
                e = k.p3;
                f = k.p4;
            }
            else {
                var l = b.h(a.c.y, j, this.yw.b, e, f);
                e = l.p3;
                f = l.p4;
            }
        }
        else {
            e = NaN;
            f = NaN;
        }
        return {
            p4: e,
            p5: f
        };
    };
    ScatterBase.prototype.abh = function (a, b, c, d, e, f) {
        var g = a.i(b, c);
        c = g.p1;
        if (f) {
            var h = a.k(c, b, this.yw.a, d, e);
            d = h.p3;
            e = h.p4;
        }
        else {
            var i = a.k(c, b, this.yw.b, d, e);
            d = i.p3;
            e = i.p4;
        }
        return {
            p2: c,
            p3: d,
            p4: e
        };
    };
    ScatterBase.prototype.abn = function (a, b) {
        if (!b.o() || this.yy == null) {
            b.db();
            return;
        }
        this.abo(a, b);
        this.abp(a, b);
    };
    ScatterBase.prototype.abo = function (a, b) {
        var e_4, _a;
        b.c7();
        var c = new ErrorBarsHelper(this, b);
        var d = new PathGeometry();
        var e = this.yy.u;
        try {
            for (var _b = __values(fromEnum(a.j.keys)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var f = _c.value;
                if (e != null && a.i.containsKey(f)) {
                    var g = a.i.item(f)[0];
                    var h = a.i.item(f)[1];
                    if (!isNaN_(g) && !isNaN_(h)) {
                        var i = a.h.item(f).c;
                        if (this.yy.o == 1 || this.yy.o == 2) {
                            c.d(d, i, g, true);
                        }
                        if (this.yy.o == 1 || this.yy.o == 3) {
                            c.d(d, i, h, false);
                        }
                    }
                }
            }
        }
        catch (e_4_1) {
            e_4 = { error: e_4_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_4)
                    throw e_4.error;
            }
        }
        b.dc(d);
    };
    ScatterBase.prototype.abp = function (a, b) {
        var e_5, _a;
        b.c8();
        var c = new ErrorBarsHelper(this, b);
        var d = new PathGeometry();
        var e = this.yy.v;
        try {
            for (var _b = __values(fromEnum(a.j.keys)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var f = _c.value;
                if (e != null && a.k.containsKey(f)) {
                    var g = a.k.item(f)[0];
                    var h = a.k.item(f)[1];
                    if (!isNaN_(g) && !isNaN_(h)) {
                        var i = a.l.item(f).c;
                        if (this.yy.p == 1 || this.yy.p == 2) {
                            c.e(d, i, g, true);
                        }
                        if (this.yy.p == 1 || this.yy.p == 3) {
                            c.e(d, i, h, false);
                        }
                    }
                }
            }
        }
        catch (e_5_1) {
            e_5 = { error: e_5_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_5)
                    throw e_5.error;
            }
        }
    };
    ScatterBase.prototype.he = function (a, b, c) {
        var d = true;
        if (!_super.prototype.he.call(this, a, b, c) || b.isEmpty || a.isEmpty || this.xAxis == null || this.yAxis == null || this.xAxis.bi == null || this.yAxis.bi == null || this.zh == null || this.zi == null || this.zh.count == 0 || this.zi.count == 0 || this.dn == null || this.dn.count != this.zh.count || this.dn.count != this.zi.count || this.xAxis.bi == null || this.yAxis.bi == null || this.xAxis.l4 == this.xAxis.l3 || this.yAxis.l4 == this.yAxis.l3) {
            d = false;
        }
        return d;
    };
    ScatterBase.prototype.getViewInfo = function (a, b) {
        if (!this.acl.isEmpty && !this.acm.isEmpty) {
            a = this.acl;
            b = this.acm;
            return {
                p0: a,
                p1: b
            };
        }
        a = this.cv.bx;
        b = this.cv.by;
        return {
            p0: a,
            p1: b
        };
    };
    ScatterBase.prototype.rs = function (a) {
        _super.prototype.rs.call(this, a);
        this.aa3();
        try {
            if (this.d5(this.cv)) {
                return;
            }
            if (this.dn != null && this.dn.count > this.z9) {
                a = false;
            }
            if (this.gf(a) && !this.gr) {
                var b = this.y2;
                if (this.d3()) {
                    if (this.dc.g) {
                        this.dc.t();
                    }
                    this.y2 = this.y4;
                    this.y4 = b;
                }
                else {
                    this.y2 = this.y0;
                    this.y0 = b;
                }
                this.abf(this.y0, this.yx);
                this.r7();
            }
            else {
                if (!this.gr) {
                    this.abf(this.y0, this.yx);
                }
                this.abq(this.y0, this.yx);
            }
        }
        finally {
            this.abr();
        }
    };
    ScatterBase.prototype.sh = function () {
        _super.prototype.sh.call(this);
        if (this.index < 0) {
            return;
        }
        this.yx.df();
    };
    ScatterBase.prototype.rt = function (a, b) {
        _super.prototype.rt.call(this, a, b);
        if (!this.g0) {
            this.cv.a4(b);
            return;
        }
        this.cv.a4(b);
        if (this.d5(this.cu)) {
            return;
        }
        var c = typeCast(ScatterBaseView.$, this.cu);
        if (!this.gs) {
            this.y3 = new ScatterFrame();
            this.abf(this.y3, c);
        }
        this.gs = false;
        this.abq(this.y3, c);
        this.g0 = false;
    };
    ScatterBase.prototype.rq = function (a, b, c, d, e) {
        _super.prototype.rq.call(this, a, b, c, d, e);
        var f = this.ba.a$j.item(d);
        var g = f;
        f.a3(c);
        if (this.d5(g)) {
            return;
        }
        if (this.yz == null) {
            this.yz = new ScatterFrame();
        }
        this.abf(this.yz, g);
        this.abq(this.yz, g);
    };
    ScatterBase.prototype.abm = function (a, b) {
        var e_6, _a, e_7, _b;
        var c = new List$1(Base.$, 0);
        try {
            for (var _c = __values(fromEnum(b.c)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var d = _d.value;
                if (!a.containsKey(d)) {
                    c.add1(d);
                }
            }
        }
        catch (e_6_1) {
            e_6 = { error: e_6_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_6)
                    throw e_6.error;
            }
        }
        try {
            for (var _e = __values(fromEnum(c)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var e = _f.value;
                b.h(e);
            }
        }
        catch (e_7_1) {
            e_7 = { error: e_7_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_7)
                    throw e_7.error;
            }
        }
    };
    ScatterBase.prototype.yo = function (a, b, c, d, e) {
        var _this = this;
        if (c == null || c.length != this.yw.c.count) {
            c = new Array(this.yw.c.count);
            for (var f = 0; f < this.yw.c.count; f++) {
                c[f] = PointUtil.create();
            }
        }
        var g = this.getEffectiveViewport1(a);
        var h = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.xAxis.ch, g);
            $ret.c = _this.getEffectiveViewport1(a);
            return $ret;
        })());
        var i = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yAxis.ch, g);
            $ret.c = _this.getEffectiveViewport1(a);
            $ret.b = _this.iy();
            return $ret;
        })());
        var j = this.wj(a);
        var k = this.yw.a.eo(j.left, h);
        var l = this.yw.a.eo(j.right, h);
        var m = this.yw.b.eo(j.bottom, i);
        var n = this.yw.b.eo(j.top, i);
        if (this.yw.d) {
            var o = k;
            k = l;
            l = o;
        }
        if (this.yw.e) {
            var p = m;
            m = n;
            n = p;
        }
        var q = this.yw;
        var r = q.a;
        var s = q.b;
        var t;
        var u;
        var v = this.zh;
        var w = this.zi;
        var x = false;
        var y = 0;
        var z = 0;
        if (this.db != null) {
            if (this.db.f4()) {
                var aa = this.db.m4(d, h);
                x = aa.c;
                y = aa.d;
                z = r.eo(0 + y, h);
                var ab = r.eo(0, h);
                z -= ab;
            }
        }
        for (var ac = 0; ac < this.yw.c.count; ac++) {
            t = v.item(ac);
            u = w.item(ac);
            var ad = 1E-10;
            if (x) {
                if (t >= k - ad && t <= l + ad && u >= m - ad && u <= n + ad) {
                    c[ac].x = r.el(t, h);
                    c[ac].y = s.el(u, i);
                }
                else {
                    if ((t + z) >= k && (t + z) <= l && u >= m && u <= n) {
                        c[ac].x = r.el(t, h) + y;
                        c[ac].y = s.el(u, i);
                    }
                    else {
                        c[ac].x = NaN;
                        c[ac].y = NaN;
                    }
                }
            }
            else {
                if (t >= k - ad && t <= l + ad && u >= m - ad && u <= n + ad) {
                    c[ac].x = r.el(t, h);
                    c[ac].y = s.el(u, i);
                }
                else {
                    c[ac].x = NaN;
                    c[ac].y = NaN;
                }
            }
        }
        return c;
    };
    ScatterBase.prototype.zx = function (a, b) {
        var e_8, _a;
        if (b == null) {
            b = new List$1(Number_$type, 2, a.e);
        }
        else {
            b.clear();
        }
        var c = 0;
        var d = this.dn;
        try {
            for (var _b = __values(fromEnum(a.c)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var e = _c.value;
                b.add(d.indexOf(e));
            }
        }
        catch (e_8_1) {
            e_8 = { error: e_8_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_8)
                    throw e_8.error;
            }
        }
        return b;
    };
    ScatterBase.prototype.pz = function (a) {
        _super.prototype.pz.call(this, a);
        this.bc.exportTrendlineData(this, a, this.yx.c2.g$j);
    };
    ScatterBase.prototype.iy = function () {
        return Series.iz(this.zi, null);
    };
    ScatterBase.prototype.getDataLegendSeriesContextAt = function (a, b) {
        var c = { $type: Point_$type, x: a, y: b };
        var d = this.kn(c);
        var e = d == null ? NaN : this.ix(d, "XMemberPath");
        var f = d == null ? NaN : this.ix(d, "YMemberPath");
        var g = new DataLegendSeriesValueInfo();
        g.l = e;
        g.a = 13;
        g.q = "X";
        g.r = this.aat != null ? this.aat : "X";
        g.s = this.aap != null ? this.aap : "X:";
        g.t = this.aar != null ? this.aar : "";
        var h = new DataLegendSeriesValueInfo();
        h.l = f;
        h.a = 14;
        h.q = "Y";
        h.r = this.aa1 != null ? this.aa1 : "Y";
        h.s = this.aax != null ? this.aax : "Y:";
        h.t = this.aaz != null ? this.aaz : "";
        var i = new DataLegendSeriesContext();
        i.e = ((function () {
            var $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
            $ret.add(g);
            $ret.add(h);
            return $ret;
        })());
        i.h = this.mc();
        return i;
    };
    ScatterBase.prototype.abs = function () {
        if (this.zj) {
            this.zy = 0;
            this.zz = 0;
            for (var a = 0; a < this.zh.count; a++) {
                this.zy += this.zh.item(a);
                this.zz += this.zi.item(a);
            }
            this.z0 = this.zy / this.zh.count;
            this.z1 = this.zz / this.zi.count;
            this.zj = false;
        }
    };
    ScatterBase.prototype.i8 = function (a, b) {
        if (b == null || !b.cj) {
            return NaN;
        }
        if (b != this.xAxis && b != this.yAxis) {
            return NaN;
        }
        var c = this.aq(a);
        if (b.cp) {
            return c[1];
        }
        else {
            return c[0];
        }
    };
    ScatterBase.prototype.aq = function (a) {
        var b = NaN;
        var c = NaN;
        switch (a) {
            case 5:
            case 2:
                b = this.zh.maximum;
                c = this.zi.maximum;
                break;
            case 4:
            case 1:
                b = this.zh.minimum;
                c = this.zi.minimum;
                break;
            case 6:
            case 3:
                if (this.zj) {
                    this.abs();
                }
                b = this.z0;
                c = this.z1;
                break;
        }
        return [b, c];
    };
    ScatterBase.prototype.wg = function (a) {
        var b = this.getEffectiveViewport1(this.cv);
        var c = new ScalerParams(0, this.cv.by, this.cv.bx, this.xAxis.ch, b);
        c.c = b;
        var d = isNaN_(a[0]) ? NaN : this.xAxis.el(a[0], c);
        c = new ScalerParams(0, this.cv.by, this.cv.bx, this.yAxis.ch, b);
        c.b = this.iy();
        c.c = b;
        var e = isNaN_(a[1]) ? NaN : this.yAxis.el(a[1], c);
        return { $type: Point_$type, x: d, y: e };
    };
    ScatterBase.prototype.wf = function (a) {
        var b = this.aq(a);
        return this.wg(b);
    };
    ScatterBase.$t = markType(ScatterBase, 'ScatterBase', MarkerSeries.$, [ISupportsErrorBars_$type, ITrendLineSeries_$type]);
    ScatterBase.abt = DependencyProperty.i("ActualTrendLineBrush", Brush.$, ScatterBase.$, new PropertyMetadata(2, null, function (a, b) { return a.raisePropertyChanged("ActualTrendLineBrush", b.oldValue, b.newValue); }));
    ScatterBase.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterBase, 'raisePropertyChanged', ['ErrorBarSettings:yy:abu', [ScatterErrorBarSettings.$, null], 'HighlightedXMemberPath:aad:abv', [2, null], 'HighlightedYMemberPath:aaf:abw', [2, null], 'IsCustomScatterMarkerStyleAllowed:zo:abx', [0, false], 'IsCustomScatterStyleAllowed:zp:aby', [0, false], 'ItemSearchMode:y8:abz', [ScatterItemSearchMode_$type, enumGetBox(ScatterItemSearchMode_$type, 0)], 'ItemSearchThreshold:z8:ab0', [1, 10000], 'MarkerCollisionAvoidance:yt:ab1', [CollisionAvoidanceType_$type, enumGetBox(CollisionAvoidanceType_$type, 0)], 'MaximumMarkers:z9:ab2', [1, 400], 'XAxis::aca', [NumericXAxis.$, null], 'XMemberAsLegendLabel:aap:acb', [2, null], 'XMemberAsLegendUnit:aar:acc', [2, null], 'XMemberPath:aat:acd', [2, null], 'YAxis::ace', [NumericYAxis.$, null], 'YMemberAsLegendLabel:aax:acf', [2, null], 'YMemberAsLegendUnit:aaz:acg', [2, null], 'YMemberPath:aa1:ach', [2, null], 'TrendLineType::ab8', [TrendLineType_$type, enumGetBox(TrendLineType_$type, 0)], 'TrendLineThickness::ab7', [1, 1.5], 'TrendLinePeriod::ab6', [1, 7], 'TrendLineBrush::ab3', [Brush.$, null], 'TrendLineZIndex:aaa:ab9', [1, 1001], 'TrendLineDashArray:acj:ab4', [DoubleCollection.$], 'TrendLineDashCap:ack:ab5', [PenLineCap_$type, enumGetBox(PenLineCap_$type, 0)]]);
    return ScatterBase;
}(MarkerSeries));
export { ScatterBase };
/**
 * @hidden
 */
var ScatterSeriesRenderManager = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterSeriesRenderManager, _super);
    function ScatterSeriesRenderManager() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.aw = null;
        _this.ax = null;
        return _this;
    }
    ScatterSeriesRenderManager.prototype.c = function () {
        this.aw = new AssigningScatterMarkerStyleEventArgs();
        return this.aw;
    };
    ScatterSeriesRenderManager.prototype.a = function () {
        this.ax = new AssigningScatterStyleEventArgs();
        return this.ax;
    };
    ScatterSeriesRenderManager.prototype.ay = function (a, b, c) {
        this.w(a, b, c);
    };
    ScatterSeriesRenderManager.prototype.az = function (a, b, c) {
        this.x(a, b, c);
    };
    ScatterSeriesRenderManager.prototype.a0 = function (a, b, c, d) {
        if (b == -1) {
            a.e = false;
            a.z = 0;
            a.y = c - 1;
        }
        else {
            if (d) {
                a.z = b;
                a.y = b;
            }
            else {
                a.e = false;
                a.z = b;
                a.y = b;
            }
        }
    };
    ScatterSeriesRenderManager.$t = markType(ScatterSeriesRenderManager, 'ScatterSeriesRenderManager', SeriesRenderManager.$);
    return ScatterSeriesRenderManager;
}(SeriesRenderManager));
export { ScatterSeriesRenderManager };
/**
 * @hidden
 */
var ScatterBaseView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterBaseView, _super);
    function ScatterBaseView(a) {
        var _this = _super.call(this, a) || this;
        _this.c0 = null;
        _this.cy = null;
        _this.c1 = null;
        _this.c6 = null;
        _this.dh = null;
        _this.di = null;
        _this.c2 = null;
        _this.dg = null;
        _this.c3 = null;
        _this.c1 = a;
        _this.c3 = new HashPool$2(Base.$, Marker.$);
        _this.cm(_this.c3);
        _this.c2 = new ScatterTrendLineManager();
        return _this;
    }
    ScatterBaseView.prototype.au = function () {
        _super.prototype.au.call(this);
        this.c0 = this.cz();
        this.dh = new Path();
        this.di = new Path();
        if (!this.q) {
            this.c1.z9 = 2000;
        }
    };
    ScatterBaseView.prototype.dj = function (a) {
        if (this.dg == null) {
            this.dg = new DataTemplateMeasureInfo();
            this.dg.context = this.k.h();
        }
        this.dg.width = a.width;
        this.dg.height = a.height;
        this.dg.data = a.content;
        var b = a.ah;
        if (b.measure != null) {
            b.measure(this.dg);
        }
        return new Size(1, this.dg.width, this.dg.height);
    };
    ScatterBaseView.prototype.cz = function () {
        var _this = this;
        var a = new NumericMarkerManager(1, function (b) { return _this.c3.item(b); }, function (b) { return _this.c1.yw.c.item(b); }, runOn(this, this.de), runOn(this, this.cw), runOn(this, this.c4), function () { return _this.c1.yt; });
        a.q = runOn(this, this.dj);
        return a;
    };
    ScatterBaseView.prototype.de = function (a) {
        this.c1.abm(a, this.c3);
    };
    ScatterBaseView.prototype.c5 = function () {
        return this.b8;
    };
    ScatterBaseView.prototype.cx = function () {
        return this.cy;
    };
    ScatterBaseView.prototype.cw = function () {
        this.cy = this.c1.yo(this, this.c3, this.cy, this.by, this.bx);
        return this.cy;
    };
    ScatterBaseView.prototype.c4 = function () {
        this.c6 = this.c1.zx(this.c3, this.c6);
        return this.c6;
    };
    ScatterBaseView.prototype.c7 = function () {
    };
    ScatterBaseView.prototype.dc = function (a) {
    };
    ScatterBaseView.prototype.c8 = function () {
    };
    ScatterBaseView.prototype.dd = function (a) {
    };
    ScatterBaseView.prototype.df = function () {
        if (this.c1.trendLineBrush != null) {
            this.c1.aci = this.c1.trendLineBrush;
        }
        else {
            this.c1.aci = this.c1.u9;
        }
    };
    ScatterBaseView.prototype.cj = function (a) {
        this.c3.f(a);
    };
    ScatterBaseView.prototype.da = function (a) {
        this.c3.g(a);
    };
    ScatterBaseView.prototype.db = function () {
    };
    ScatterBaseView.prototype.a7 = function (a, b) {
        _super.prototype.a7.call(this, a, b);
        if (a.d) {
            if (this.c1.trendLineType != 0 && !b) {
                var c = this.c2.g$j;
                c.ad = this.c1.trendLineThickness;
                c._stroke = this.c1.aci;
                c.ai = this.c1.acj;
                c.aj = this.c1.ack;
                a.v(c);
            }
        }
    };
    ScatterBaseView.prototype.c9 = function (a) {
        if (a) {
            this.db();
            this.c3.clear();
        }
        this.c2.g$d();
        this.as();
    };
    ScatterBaseView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span";
        var b = this.e.ba.a$u(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.c1.title + "</span><br/><span>" + "x: </span><span class='ui-priority-primary'>${item." + this.c1.aat + "}</span><br/><span>" + "y: </span><span class='ui-priority-primary'>${item." + this.c1.aa1 + "}</span></div>";
        return a;
    };
    ScatterBaseView.$t = markType(ScatterBaseView, 'ScatterBaseView', MarkerSeriesView.$);
    return ScatterBaseView;
}(MarkerSeriesView));
export { ScatterBaseView };
