// igniteui-react-charts/esm5/lib/ScatterPolygonSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { ShapeSeriesBase } from "./ShapeSeriesBase";
import { ShapeAxisInfoCache } from "./ShapeAxisInfoCache";
import { StyleSelector } from "./StyleSelector";
import { DependencyProperty } from "igniteui-react-core";
import { Style } from "igniteui-react-core";
import { MarkerType_$type } from "./MarkerType";
import { DataTemplate } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
import { CollisionAvoidanceType_$type } from "./CollisionAvoidanceType";
import { MarkerOutlineMode_$type } from "./MarkerOutlineMode";
import { MarkerFillMode_$type } from "./MarkerFillMode";
import { Rect } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { Base, runOn, enumGetBox, Number_$type, typeCast, fromEnum, markType, markDep, PointUtil } from "igniteui-react-core";
import { Dictionary$2 } from "igniteui-react-core";
import { OwnedPoint } from "./OwnedPoint";
import { ShapeSeriesViewBase } from "./ShapeSeriesViewBase";
import { Marker } from "./Marker";
import { MarkerSeries } from "./MarkerSeries";
import { BrushUtil } from "igniteui-react-core";
import { DataContext } from "igniteui-react-core";
import { HashPool$2 } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { RectUtil } from "igniteui-react-core";
import { ChartSelection } from "./ChartSelection";
import { PropertyMetadata } from "igniteui-react-core";
import { DeviceUtils } from "igniteui-react-core";
import { isNaN_, isInfinity } from "igniteui-react-core";
import { NumericMarkerManager } from "./NumericMarkerManager";
import { DataTemplateMeasureInfo } from "igniteui-react-core";
import { Size } from "igniteui-react-core";
import { Defaults } from "./Defaults";
import { DataTemplatePassInfo } from "igniteui-react-core";
import { DataTemplateRenderInfo } from "igniteui-react-core";
import { MarkerVisualData } from "./MarkerVisualData";
import { stringIsNullOrEmpty } from "igniteui-react-core";
/**
 * @hidden
 */
var ScatterPolygonSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolygonSeries, _super);
    function ScatterPolygonSeries() {
        var _this = _super.call(this) || this;
        _this.aal = null;
        _this.aai = new ScalerParams(1, ScatterPolygonSeries.abr, ScatterPolygonSeries.abr, false);
        _this.aaj = new ScalerParams(1, ScatterPolygonSeries.abr, ScatterPolygonSeries.abr, false);
        _this.aa7 = null;
        _this.aad = 1;
        _this.z9 = 1;
        _this.ab = ScatterPolygonSeries.$;
        return _this;
    }
    ScatterPolygonSeries.prototype.bj = function () {
        return new ScatterPolygonSeries();
    };
    ScatterPolygonSeries.prototype.get_fq = function () {
        return true;
    };
    ScatterPolygonSeries.prototype.get_fl = function () {
        return true;
    };
    ScatterPolygonSeries.prototype.yf = function () {
        return this.aan();
    };
    ScatterPolygonSeries.prototype.zn = function (a) {
        var _this = this;
        var b = a;
        b.cy(this.abt);
        b.cz(this.aam);
        _super.prototype.zn.call(this, a);
        if (this.d5(a)) {
            return;
        }
        var c, d;
        var e = a.getViewInfo(c, d);
        c = e.p0;
        d = e.p1;
        var f = this.getEffectiveViewport1(a);
        this.aai = new ScalerParams(1, d, c, this.xe.ch);
        this.aai.c = f;
        this.aaj = new ScalerParams(1, d, c, this.xg.ch);
        this.aaj.c = f;
        if (c.width < 1 || c.height < 1) {
            return;
        }
        this.aal = ((function () {
            var $ret = new ShapeAxisInfoCache();
            $ret.a = _this.xe;
            $ret.b = _this.xg;
            $ret.e = _this.xe.ch;
            $ret.f = _this.xg.ch;
            $ret.d = _this.dn;
            $ret.c = _this.xw;
            return $ret;
        })());
        if (this.aan()) {
            var g = new Dictionary$2(Base.$, OwnedPoint.$, 0);
            b.dg.l(g, 400, d, c, this.id);
            b.dg.j(g, false);
            if (this.yd()) {
                this.aa4(a);
            }
        }
    };
    ScatterPolygonSeries.prototype.aa4 = function (a) {
        var _this = this;
        a.dk = true;
        this.xl.ay(this, this.yd(), runOn(this, this.w4));
        var b = this.xl.d;
        var c = b != null;
        var d = this.getEffectiveViewport1(a);
        var e = this.dn == null ? 0 : this.dn.count;
        a.dw(function (f) {
            if (f._visibility == 0) {
                var g = f.ak;
                var h = f.content;
                if (h == null) {
                    h = new DataContext();
                    f.content = h;
                }
                if (c) {
                    _this.zi(g, e, a.q);
                }
                _this.xl.ac(f, h);
            }
        });
    };
    ScatterPolygonSeries.prototype.aan = function () {
        return this.aa7 != null && ((this.aah != 1 && this.aah != 0) || this.aa8 != null);
    };
    ScatterPolygonSeries.prototype.cs = function () {
        return new ScatterPolygonSeriesView(this);
    };
    ScatterPolygonSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.aak = a;
    };
    ScatterPolygonSeries.prototype.get_hasMarkers = function () {
        return true;
    };
    ScatterPolygonSeries.prototype.get_el = function () {
        return this.aah != 13 && this.aah != 1;
    };
    ScatterPolygonSeries.prototype.vl = function () {
        return this.abn;
    };
    ScatterPolygonSeries.prototype.vm = function () {
        return this.abo;
    };
    ScatterPolygonSeries.prototype.sz = function () {
        return this.aa6;
    };
    ScatterPolygonSeries.prototype.ma = function () {
        return this.mb(this, this.aa7);
    };
    ScatterPolygonSeries.prototype.io = function () {
        return this.aap;
    };
    Object.defineProperty(ScatterPolygonSeries.prototype, "aae", {
        get: function () {
            return this.aad;
        },
        set: function (a) {
            this.aad = a;
        },
        enumerable: false,
        configurable: true
    });
    ScatterPolygonSeries.prototype.aag = function () {
        if (this.aaf == 0) {
            return 1;
        }
        return this.aaf;
    };
    ScatterPolygonSeries.prototype.sl = function (a) {
        _super.prototype.sl.call(this, a);
        this.aaf = a;
    };
    Object.defineProperty(ScatterPolygonSeries.prototype, "aaa", {
        get: function () {
            return this.z9;
        },
        set: function (a) {
            this.z9 = a;
        },
        enumerable: false,
        configurable: true
    });
    ScatterPolygonSeries.prototype.aac = function () {
        if (this.aab == 0) {
            return 2;
        }
        return this.aab;
    };
    ScatterPolygonSeries.prototype.sk = function (a) {
        _super.prototype.sk.call(this, a);
        this.aab = a;
    };
    ScatterPolygonSeries.prototype.sm = function (a) {
        _super.prototype.sm.call(this, a);
        this.aap = a;
    };
    ScatterPolygonSeries.prototype.sh = function () {
        _super.prototype.sh.call(this);
        if (this.index < 0) {
            return;
        }
        if (this.aa8 != null) {
            this.aak.dv();
            this.aak.ds();
        }
        else {
            var a = MarkerSeries.xb(this, this.aah);
            var b = MarkerSeries.xo(a);
            if (b == null) {
                this.aa6 = MarkerSeries.x3;
            }
            else {
                this.aak.dp(b);
            }
        }
        if (this.abp != null && this.aah != 13) {
            this.r8("ActualMarkerBrush");
            var c = this.abn;
            this.aak.dt();
            this.aak.dq();
            var d = this.abn;
            var e = this.fh("ActualMarkerBrush", this.hv.f > 0 ? this.hv.d() : null, c, d);
            this.sa("ActualMarkerBrush", e);
        }
        else {
            if (this.aah == 13) {
                this.abn = this.cv.bt();
            }
            else {
                if (this.aaa != 2) {
                    if (this.db != null) {
                        var f = this.db.ve(this.index);
                        if (!isNaN_(this.ic) && this.ic != 1) {
                            f = BrushUtil.q(f, this.ic);
                        }
                        this.abn = f;
                    }
                    else {
                        this.abn = null;
                    }
                }
            }
        }
        if (this.abq != null) {
            this.aak.du();
            this.aak.dr();
        }
        else {
            if (this.aah == 13) {
                this.abo = this.cv.bt();
            }
            else {
                if (this.aaf == 2) {
                    this.abo = this.abn;
                }
                else {
                    this.abo = this.db == null ? null : this.db.vf(this.index);
                }
            }
        }
        if (this.aaa == 2 && this.abp == null && this.aah != 13) {
            if (this.abo != null) {
                var g = this.abo;
                if (!isNaN_(this.ic) && this.ic != 1) {
                    g = BrushUtil.q(g, this.ic);
                }
                this.abn = g;
            }
            else {
                this.abn = null;
            }
        }
    };
    ScatterPolygonSeries.prototype.qs = function () {
        _super.prototype.qs.call(this);
        this.aae = this.aag();
        this.aaa = this.aac();
    };
    ScatterPolygonSeries.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case ScatterPolygonSeries.$$p[3]:
            case ScatterPolygonSeries.$$p[11]:
            case ScatterPolygonSeries.$$p[7]:
            case ScatterPolygonSeries.$$p[9]:
            case ScatterPolygonSeries.$$p[10]:
                this.rf(b);
                this.q3();
                break;
            case ScatterPolygonSeries.$$p[6]:
                this.aae = this.aag();
                this.rf(null);
                this.q3();
                break;
            case ScatterPolygonSeries.$$p[5]:
                this.aaa = this.aac();
                this.rf(null);
                this.q3();
                break;
            case ScatterPolygonSeries.$$p[2]:
                this.aa7 = d;
                if (c == MarkerSeries.x3 || d == MarkerSeries.x3 || (c == null || d != null)) {
                    this.aak.dx();
                    this.rr(false);
                }
                this.cv.bl();
                if (this.toolTip != null && this.fi) {
                    this.ba.a$as(this, this.toolTip);
                }
                this.rr(false);
                this.ql();
                break;
            case ScatterPolygonSeries.$$p[12]:
                this.aak.cy(this.abt);
                this.rr(false);
                this.ql();
                break;
            case ScatterPolygonSeries.$$p[13]:
                this.aak.cz(this.aam);
                this.rr(false);
                this.ql();
                break;
            case ScatterPolygonSeries.$$p[4]:
                this.rr(false);
                this.ql();
                break;
            case ScatterPolygonSeries.$$p[0]:
            case ScatterPolygonSeries.$$p[1]:
                if (this.aan()) {
                    this.rr(false);
                }
                break;
        }
    };
    ScatterPolygonSeries.prototype.a2 = function (a, b) {
        var c = this.aak.b4(a);
        var d = null;
        if (c != null) {
            d = c.content;
        }
        if (d != null) {
            return d;
        }
        return _super.prototype.a2.call(this, a, b);
    };
    ScatterPolygonSeries.prototype.aa5 = function (a, b) {
        var e_1, _a, e_2, _b;
        var c = new List$1(Base.$, 0);
        try {
            for (var _c = __values(fromEnum(b.c)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var d = _d.value;
                if (!a.containsKey(d)) {
                    c.add1(d);
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        try {
            for (var _e = __values(fromEnum(c)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var e = _f.value;
                b.h(e);
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
    };
    ScatterPolygonSeries.prototype.eb = function (a, b, c, d, e) {
        var _this = this;
        if (!this.hasMarkers) {
            return false;
        }
        var f = this.xe;
        var g = new ScalerParams(1, this.db.v4, this.ws, f.ch);
        g.c = this.getEffectiveViewport1(this.cv);
        var h = this.xg;
        var i = new ScalerParams(1, this.db.v4, this.ws, h.ch);
        i.c = this.getEffectiveViewport1(this.cv);
        a.clear();
        this.aak.dw(function (j) {
            if (j._visibility == 0) {
                var k = new SeriesPointOfInterest();
                var l = j.n;
                var m = j.o;
                if (l < _this.ws.left || l > _this.ws.right || m < _this.ws.top || m > _this.ws.bottom) {
                    return;
                }
                k.d = f.eo(l, g);
                k.e = h.eo(m, i);
                k.b = j.content;
                k.g = _this.vg;
                k.c = j.ak;
                if (_this.g9) {
                    k.g = _this.vn(j.ak);
                }
                a.add(k);
            }
        });
        return true;
    };
    ScatterPolygonSeries.prototype.z7 = function (a, b, c, d, e) {
        var _this = this;
        if (c == null || c.length != this.aal.d.count) {
            c = new Array(this.aal.d.count);
            for (var f = 0; f < this.aal.d.count; f++) {
                c[f] = PointUtil.create();
            }
        }
        var g = ((function () {
            var $ret = new ScalerParams(1, d, e, _this.xe.ch);
            $ret.c = _this.getEffectiveViewport1(a);
            return $ret;
        })());
        var h = ((function () {
            var $ret = new ScalerParams(1, d, e, _this.xg.ch);
            $ret.c = _this.getEffectiveViewport1(a);
            return $ret;
        })());
        var i = this.wj(a);
        var j = this.aal.a.eo(i.left, g);
        var k = this.aal.a.eo(i.right, g);
        var l = this.aal.b.eo(i.bottom, h);
        var m = this.aal.b.eo(i.top, h);
        if (this.aal.e) {
            var n = j;
            j = k;
            k = n;
        }
        if (this.aal.f) {
            var o = l;
            l = m;
            m = o;
        }
        var p = this.aal;
        var q = p.a;
        var r = p.b;
        var s;
        var t;
        var u = g.d.left;
        var v = g.d.right;
        var w = h.d.top;
        var x = h.d.bottom;
        for (var y = 0; y < this.aal.d.count; y++) {
            s = 0;
            t = 0;
            var z = null;
            var aa = a.cc.tryGetValue(y, z);
            z = aa.p1;
            if (z == null) {
                c[y].x = NaN;
                c[y].y = NaN;
            }
            else {
                var ab = z.c;
                var ac = RectUtil.l(ab);
                s = ac.x;
                t = ac.y;
                if (s >= u && s <= v && t >= w && t <= x) {
                    c[y].x = s;
                    c[y].y = t;
                }
                else {
                    c[y].x = NaN;
                    c[y].y = NaN;
                }
            }
        }
        return c;
    };
    ScatterPolygonSeries.prototype.aao = function (a, b) {
        var e_3, _a;
        if (b == null) {
            b = new List$1(Number_$type, 2, a.e);
        }
        else {
            b.clear();
        }
        var c = 0;
        var d = this.dn;
        try {
            for (var _b = __values(fromEnum(a.c)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var e = _c.value;
                b.add(d.indexOf(e));
                c++;
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
        return b;
    };
    ScatterPolygonSeries.prototype.ak = function (a, b) {
        var c = new List$1(ChartSelection.$, 0);
        this.p4(a, c, b);
        return c.toArray();
    };
    ScatterPolygonSeries.prototype.p4 = function (a, b, c) {
        this.zf(a, b, c);
    };
    ScatterPolygonSeries.$t = markType(ScatterPolygonSeries, 'ScatterPolygonSeries', ShapeSeriesBase.$);
    ScatterPolygonSeries.abr = new Rect(0, 0, 0, 1, 1);
    ScatterPolygonSeries.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterPolygonSeries, 'raisePropertyChanged', ['ActualMarkerBrush:abn:aa9', [Brush.$], 'ActualMarkerOutline:abo:aba', [Brush.$], 'ActualMarkerTemplate:aa6:abb', [DataTemplate.$, null], 'MarkerBrush:abp:abc', [Brush.$], 'MarkerCollisionAvoidance:z8:abd', [CollisionAvoidanceType_$type, enumGetBox(CollisionAvoidanceType_$type, 0)], 'MarkerFillMode:aab:abe', [MarkerFillMode_$type, enumGetBox(MarkerFillMode_$type, 0)], 'MarkerOutlineMode:aaf:abf', [MarkerOutlineMode_$type, enumGetBox(MarkerOutlineMode_$type, 0)], 'MarkerOutline:abq:abg', [Brush.$], 'MarkerStyle:abs:abh', [Style.$], 'MarkerTemplate:aa8:abi', [DataTemplate.$, null], 'MarkerThickness:aap:abj', [1, DeviceUtils.g(2)], 'MarkerType:aah:abk', [MarkerType_$type, enumGetBox(MarkerType_$type, 1)], 'ShapeStyle:abt:abl', [Style.$, null], 'ShapeStyleSelector:aam:abm', [StyleSelector.$, null]]);
    return ScatterPolygonSeries;
}(ShapeSeriesBase));
export { ScatterPolygonSeries };
/**
 * @hidden
 */
var ScatterPolygonSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolygonSeriesView, _super);
    function ScatterPolygonSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.dk = false;
        _this.db = null;
        _this.dm = null;
        _this.dh = null;
        _this.df = null;
        _this.dj = false;
        _this.di = null;
        _this.dn = null;
        _this.dc = new Marker();
        _this.d4 = null;
        _this.dc = new Marker();
        _this.dc.content = new DataContext();
        _this.dh = a;
        _this.di = new HashPool$2(Base.$, Marker.$);
        _this.dn = new List$1(Marker.$, 0);
        _this.dy(_this.di);
        _this.df = new NumericMarkerManager(1, function (b) { return _this.di.item(b); }, function (b) { return _this.dh.aal.d.item(b); }, runOn(_this, _this.d2), runOn(_this, _this.da), runOn(_this, _this.dl), function () { return _this.dh.z8; });
        _this.df.q = runOn(_this, _this.d5);
        return _this;
    }
    ScatterPolygonSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.dh.abq = Defaults.am;
            this.dh.jp = Defaults.z;
            this.dh.aap = Defaults.y;
            this.dh.aah = Defaults.f;
            this.dh.aab = Defaults.b;
        }
    };
    Object.defineProperty(ScatterPolygonSeriesView.prototype, "dg", {
        get: function () {
            return this.df;
        },
        enumerable: false,
        configurable: true
    });
    ScatterPolygonSeriesView.prototype.d2 = function (a) {
        this.dh.aa5(a, this.di);
    };
    ScatterPolygonSeriesView.prototype.da = function () {
        this.db = this.dh.z7(this, this.di, this.db, this.by, this.bx);
        return this.db;
    };
    ScatterPolygonSeriesView.prototype.dl = function () {
        this.dm = this.dh.aao(this.di, this.dm);
        return this.dm;
    };
    ScatterPolygonSeriesView.prototype.c5 = function (a, b) {
        return this.c6(a, b, true);
    };
    ScatterPolygonSeriesView.prototype.cm = function (a, b, c) {
        _super.prototype.cm.call(this, a, b, c);
        ShapeSeriesViewBase.cl(this, a, b);
    };
    ScatterPolygonSeriesView.prototype.ck = function (a, b) {
        var c = a;
        if (c == null) {
            return;
        }
        c.an = b;
    };
    ScatterPolygonSeriesView.prototype.dw = function (a) {
        this.di.f(a);
    };
    ScatterPolygonSeriesView.prototype.dy = function (a) {
        a.create = runOn(this, this.de);
        a.destroy = runOn(this, this.d0);
        a.activate = runOn(this, this.dz);
        a.disactivate = runOn(this, this.d1);
    };
    ScatterPolygonSeriesView.prototype.de = function () {
        var _this = this;
        var a = new Marker();
        if (!this.dj) {
            a.content = ((function () {
                var $ret = new DataContext();
                $ret.series = _this.e;
                return $ret;
            })());
        }
        a.ah = this.e.aa6;
        this.dn.add(a);
        return a;
    };
    ScatterPolygonSeriesView.prototype.b4 = function (a) {
        var b = this.dd(a, 0);
        if (b == null) {
            b = this.dd(a, 3);
        }
        return b;
    };
    ScatterPolygonSeriesView.prototype.dd = function (a, b) {
        var c;
        var d;
        var e;
        var f;
        for (var g = this.dn.count - 1; g >= 0; g--) {
            var h = this.dn._inner[g];
            if (h._visibility == 1 || h._opacity == 0) {
                continue;
            }
            c = (h.m / 2) + b;
            d = (h.l / 2) + b;
            e = h.ai;
            f = h.aj;
            if ((h.n + e) - c <= a.x && (h.n + e) + c >= a.x && (h.o + f) - d <= a.y && (h.o + f) + d >= a.y) {
                return h;
            }
        }
        return null;
    };
    ScatterPolygonSeriesView.prototype.d0 = function (a) {
        a.content = null;
        this.dn.remove(a);
    };
    ScatterPolygonSeriesView.prototype.dz = function (a) {
        a._visibility = 0;
    };
    ScatterPolygonSeriesView.prototype.d1 = function (a) {
        if (a.content != null) {
            (typeCast(DataContext.$, a.content)).item = null;
        }
        a._visibility = 1;
    };
    ScatterPolygonSeriesView.prototype.d3 = function (a) {
        this.dj = a;
    };
    ScatterPolygonSeriesView.prototype.bj = function (a, b) {
        _super.prototype.bj.call(this, a, b);
        if (!this.dk) {
            var c = a;
            var d = c.content;
            if (d != null) {
                if (this.dh.abn != null) {
                    d.actualItemBrush = this.dh.abn;
                }
                else {
                    d.actualItemBrush = this.b6.u9;
                }
                if (this.dh.abn != null) {
                    d.outline = this.dh.abo;
                }
                else {
                    d.outline = this.b6.vd;
                }
                d.thickness = this.dh.aap;
            }
        }
    };
    ScatterPolygonSeriesView.prototype.bk = function (a, b) {
        _super.prototype.bk.call(this, a, b);
        var c = a;
        this.dc._visibility = c._visibility;
        this.dc.ah = c.ah;
        this.dc.width = c.width;
        this.dc.height = c.height;
        this.dc.m = c.m;
        this.dc.l = c.l;
        this.dc.n = c.n;
        this.dc.o = c.o;
        var d = this.bs(b);
        var e = this.dc.content;
        var f = c.content;
        e.item = f.item;
        e.series = f.series;
        e.thickness = f.thickness;
        if (e != null) {
            e.actualItemBrush = d;
            e.outline = d;
            e.thickness = this.dh.aap + 3;
        }
    };
    ScatterPolygonSeriesView.prototype.a7 = function (a, b) {
        _super.prototype.a7.call(this, a, b);
        var c = new DataTemplatePassInfo();
        c.isHitTestRender = b;
        c.context = a.h();
        c.viewportTop = this.bx.top;
        c.viewportLeft = this.bx.left;
        c.viewportWidth = this.bx.width;
        c.viewportHeight = this.bx.height;
        c.passID = "Markers";
        var d = new DataTemplateRenderInfo();
        d.passInfo = c;
        d.isHitTestRender = b;
        var e = new DataTemplateMeasureInfo();
        e.passInfo = c;
        var f = false;
        var g = a.h();
        e.context = g;
        d.context = g;
        var h = 0;
        var i = 0;
        if (this.dh.aa6 != null && this.dh.aa6.passStarting != null) {
            this.dh.aa6.passStarting(c);
        }
        for (var j = 0; j < this.dn.count; j++) {
            var k = this.dn._inner[j];
            if (k._visibility == 1) {
                continue;
            }
            this.bi(k, j, b);
            if (b) {
                k = this.dc;
            }
            if (!f) {
                e.width = k.width;
                e.height = k.height;
                e.renderOffsetX = 0;
                e.renderOffsetY = 0;
                e.renderContext = a;
                var l = k.ah;
                if (l.measure != null) {
                    e.data = k.content;
                    l.measure(e);
                    f = e.isConstant;
                    if (f) {
                        h = e.width;
                        i = e.height;
                    }
                }
                d.availableWidth = e.width;
                d.availableHeight = e.height;
                d.renderOffsetX = e.renderOffsetX;
                d.renderOffsetY = e.renderOffsetY;
                d.renderContext = a;
            }
            else {
                d.availableWidth = h;
                d.availableHeight = i;
            }
            if (!isNaN_(k.width) && !isInfinity(k.width)) {
                d.availableWidth = k.width;
            }
            if (!isNaN_(k.height) && !isInfinity(k.height)) {
                d.availableHeight = k.height;
            }
            a.q(d, k);
            k.m = d.availableWidth;
            k.l = d.availableHeight;
            k.ai = d.renderOffsetX;
            k.aj = d.renderOffsetY;
        }
        if (this.dh.aa6 != null && this.dh.aa6.passCompleted != null) {
            this.dh.aa6.passCompleted(c);
        }
    };
    ScatterPolygonSeriesView.prototype.dv = function () {
        this.dh.aa6 = null;
    };
    ScatterPolygonSeriesView.prototype.ds = function () {
        this.dh.aa6 = null;
        this.dh.aa6 = this.dh.aa8;
    };
    ScatterPolygonSeriesView.prototype.dt = function () {
        this.dh.abn = null;
    };
    ScatterPolygonSeriesView.prototype.dq = function () {
        this.dh.abn = null;
        if (!isNaN_(this.e.ic) && this.e.ic != 1) {
            var a = BrushUtil.q(this.dh.abp, this.e.ic);
            this.dh.abn = a;
        }
        else {
            this.dh.abn = this.dh.abp;
        }
    };
    ScatterPolygonSeriesView.prototype.du = function () {
        this.dh.abo = null;
    };
    ScatterPolygonSeriesView.prototype.dr = function () {
        this.dh.abo = null;
        this.dh.abo = this.dh.abq;
    };
    ScatterPolygonSeriesView.prototype.dp = function (a) {
        switch (a) {
            case "CircleMarkerTemplate":
                this.dh.aa6 = this.dh.db.r1;
                break;
            case "TriangleMarkerTemplate":
                this.dh.aa6 = this.dh.db.sa;
                break;
            case "PyramidMarkerTemplate":
                this.dh.aa6 = this.dh.db.r7;
                break;
            case "SquareMarkerTemplate":
                this.dh.aa6 = this.dh.db.r8;
                break;
            case "DiamondMarkerTemplate":
                this.dh.aa6 = this.dh.db.r2;
                break;
            case "PentagonMarkerTemplate":
                this.dh.aa6 = this.dh.db.r5;
                break;
            case "HexagonMarkerTemplate":
                this.dh.aa6 = this.dh.db.r3;
                break;
            case "TetragramMarkerTemplate":
                this.dh.aa6 = this.dh.db.r5;
                break;
            case "PentagramMarkerTemplate":
                this.dh.aa6 = this.dh.db.r6;
                break;
            case "HexagramMarkerTemplate":
                this.dh.aa6 = this.dh.db.r4;
                break;
        }
    };
    ScatterPolygonSeriesView.prototype.dx = function () {
        var e_4, _a;
        try {
            for (var _b = __values(fromEnum(this.dn)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var a = _c.value;
                a.ah = this.dh.aa6;
            }
        }
        catch (e_4_1) {
            e_4 = { error: e_4_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_4)
                    throw e_4.error;
            }
        }
        this.as();
    };
    ScatterPolygonSeriesView.prototype.d5 = function (a) {
        if (this.d4 == null) {
            this.d4 = new DataTemplateMeasureInfo();
            this.d4.context = this.k.h();
        }
        this.d4.width = a.width;
        this.d4.height = a.height;
        this.d4.data = a.content;
        var b = a.ah;
        if (b.measure != null) {
            b.measure(this.d4);
        }
        return new Size(1, this.d4.width, this.d4.height);
    };
    ScatterPolygonSeriesView.prototype.an = function (a) {
        var e_5, _a;
        var b = a;
        _super.prototype.an.call(this, b);
        try {
            for (var _b = __values(fromEnum(this.di.c)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                var d = this.di.item(c);
                if (d._visibility != 1) {
                    var e = typeCast(ScatterPolygonSeries.$, this.e);
                    var f = this.e.bc.getMarkerVisualData(e, d, e.aa6);
                    b.markerShapes.add(typeCast(MarkerVisualData.$, f));
                }
            }
        }
        catch (e_5_1) {
            e_5 = { error: e_5_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_5)
                    throw e_5.error;
            }
        }
    };
    ScatterPolygonSeriesView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span";
        var b = this.e.ba.a$u(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.dh.title + "</span><br/>";
        a += "</div>";
        return a;
    };
    ScatterPolygonSeriesView.$t = markType(ScatterPolygonSeriesView, 'ScatterPolygonSeriesView', ShapeSeriesViewBase.$);
    return ScatterPolygonSeriesView;
}(ShapeSeriesViewBase));
export { ScatterPolygonSeriesView };
