// igniteui-react-charts/esm5/lib/ScatterLineSeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ScatterBase } from "./ScatterBase";
import { UnknownValuePlotting_$type } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { Point_$type, typeCast, runOn, Number_$type, enumGetBox, markType, markDep } from "igniteui-react-core";
import { OwnedPoint } from "./OwnedPoint";
import { ScatterFrame } from "./ScatterFrame";
import { ScatterLineSeriesView } from "./ScatterLineSeriesView";
import { Defaults } from "./Defaults";
import { PathGeometry } from "igniteui-react-core";
import { Rect } from "igniteui-react-core";
import { ScalerParams } from "./ScalerParams";
import { Clipper } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { RectUtil } from "igniteui-react-core";
import { PolyLineSegment } from "igniteui-react-core";
import { PathFigure } from "igniteui-react-core";
import { PathFigureCollection } from "igniteui-react-core";
import { Flattener } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { boxArray$1, arrayShallowClone } from "igniteui-react-core";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
var ScatterLineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterLineSeries, _super);
    function ScatterLineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = ScatterLineSeries.$;
        _this.y2 = new ScatterFrame();
        _this.y4 = new ScatterFrame();
        _this.y0 = new ScatterFrame();
        var a = function (b, c, d) {
            var e = _this.dn.indexOf(b.a);
            if (e == -1 || e == 0) {
                return b;
            }
            var f = _this.dn.item(e - 1);
            var g;
            if (!((function () { var h = c.g.tryGetValue(f, g); g = h.p1; return h.ret; })())) {
                return b;
            }
            return g;
        };
        _this.y2.o = a;
        _this.y4.o = a;
        _this.y0.o = a;
        return _this;
    }
    ScatterLineSeries.prototype.bj = function () {
        return new ScatterLineSeries();
    };
    ScatterLineSeries.prototype.get_fk = function () {
        return true;
    };
    ScatterLineSeries.prototype.cs = function () {
        return new ScatterLineSeriesView(this);
    };
    ScatterLineSeries.prototype.zw = function () {
        return true;
    };
    ScatterLineSeries.prototype.jk = function () {
        if (!isNaN_(this.jj)) {
            return _super.prototype.jk.call(this);
        }
        return Defaults.x;
    };
    ScatterLineSeries.prototype.gw = function (a, b) {
        if (this.xh(a, b)) {
            return true;
        }
        if (this.aco(a, b)) {
            return true;
        }
        return false;
    };
    ScatterLineSeries.prototype.aco = function (a, b) {
        var c = this.cv;
        var d = c.dl;
        return this.gx(a, b, d.an);
    };
    ScatterLineSeries.prototype.abf = function (a, b) {
        var _this = this;
        _super.prototype.abf.call(this, a, b);
        a.m.clear();
        a.u.clear();
        var c = b.by;
        var d = b.bx;
        var e = this.getEffectiveViewport1(b);
        var f = this.id * this.id;
        var g = 0;
        if (this.zh != null) {
            g = this.zh.count;
        }
        if (this.zi != null) {
            g = Math.min(this.zi.count, g);
        }
        var h = new ScalerParams(0, c, d, this.yw.d, e);
        var i = ((function () {
            var $ret = new ScalerParams(0, c, d, _this.yw.e, e);
            $ret.b = _this.iy();
            return $ret;
        })());
        var j = arrayShallowClone(this.zh.asArray());
        var k = arrayShallowClone(this.zi.asArray());
        this.xAxis.hq(boxArray$1(j), 0, j.length, h);
        this.yAxis.hq(boxArray$1(k), 0, k.length, i);
        var l = d.top - 10;
        var m = d.bottom + 10;
        var n = d.left - 10;
        var o = d.right + 10;
        var p = ((function () {
            var $ret = new Clipper(1, n, m, o, l, false);
            $ret.i = a.m;
            return $ret;
        })());
        for (var q = 0; q < g;) {
            var r = q;
            ++q;
            if (g > this.z9) {
                while (q < g) {
                    var s = j[q] - j[r];
                    var t = k[q] - k[r];
                    var u = s * s + t * t;
                    if (isNaN_(u) || u >= f) {
                        break;
                    }
                    ++q;
                }
                p.j(this.acv(j, k, r, q - 1));
            }
            else {
                var v = new OwnedPoint();
                v.c = { $type: Point_$type, x: j[r], y: k[r] };
                v.a = this.dn.item(r);
                if (!a.u.containsKey(v.a)) {
                    a.u.addItem(v.a, v);
                }
            }
        }
        if (g > this.z9) {
            p.i = null;
        }
    };
    ScatterLineSeries.prototype.acv = function (a, b, c, d) {
        if (c == d) {
            return { $type: Point_$type, x: a[c], y: b[c] };
        }
        var e = 0;
        var f = 0;
        var g = (d - c + 1);
        for (var h = c; h <= d; ++h) {
            e += a[h];
            f += b[h];
        }
        return { $type: Point_$type, x: e / g, y: f / g };
    };
    ScatterLineSeries.prototype.get_zq = function () {
        return true;
    };
    ScatterLineSeries.prototype.abq = function (a, b) {
        b.c9(false);
        _super.prototype.abq.call(this, a, b);
        var c = this.getEffectiveViewport1(b);
        var d = new Rect(0, b.bx.left, b.bx.top, b.bx.width, b.bx.height);
        d = RectUtil.t(d, this.ie, this.ie);
        this.abj(a, new Clipper(0, d, false));
        var e = typeCast(ScatterLineSeriesView.$, b);
        this.y9.az(this, this.zv(), runOn(this, this.yp));
        var f = false;
        var g = this.y9.b;
        if (g != null) {
            f = true;
        }
        if (f) {
            this.abb(-1, this.zh.count, b.q);
        }
        this.y9.ad(e.dl, true, false, true, false);
        this.act(e.dl, a.m.count, function (h) { return a.m._inner[h].x; }, function (h) { return a.m._inner[h].y; }, this.acn, this.id);
    };
    ScatterLineSeries.prototype.act = function (a, b, c, d, e, f) {
        var g = new PathGeometry();
        a.an = g;
        g.c = new PathFigureCollection();
        var h = new List$1(PolyLineSegment.$, 0);
        var i = 0;
        for (var j = 0; j < b; j++) {
            if (isNaN_(c(j)) || isNaN_(d(j))) {
                var k = j - i;
                if (k > 0) {
                    if (e == 1 || h.count == 0) {
                        var l = new PolyLineSegment();
                        h.add(l);
                    }
                    this.acs(h._inner[h.count - 1]._points, i, j - 1, c, d, f);
                }
                i = j + 1;
            }
        }
        if (e == 1 || h.count == 0) {
            var m = new PolyLineSegment();
            h.add(m);
        }
        this.acs(h._inner[h.count - 1]._points, i, b - 1, c, d, f);
        var _loop_1 = function (n) {
            var o = h._inner[n];
            if (o._points.count > 0) {
                var p = ((function () {
                    var $ret = new PathFigure();
                    $ret.startPoint = o._points._inner[0];
                    return $ret;
                })());
                p._segments.add(o);
                g.c.add(p);
            }
        };
        for (var n = 0; n < h.count; n++) {
            _loop_1(n);
        }
    };
    ScatterLineSeries.prototype.acp = function (a, b, c, d, e, f) {
        var g = new List$1(Number_$type, 0);
        var h = d;
        var i = e;
        var j = i - h + 1;
        while (j > 0) {
            if (j <= 512) {
                Flattener.c(g, b, c, h, i, f);
                h = i + 1;
            }
            else {
                var k = h + 512 - 1;
                Flattener.c(g, b, c, h, k, f);
                h = k + 1;
            }
            j = i - h + 1;
        }
        return g;
    };
    ScatterLineSeries.prototype.acs = function (a, b, c, d, e, f) {
        if (c > -1) {
            var g = !isNaN_(this.jj) ? f : 1;
            var h = this.acp(new List$1(Number_$type, 0), d, e, b, c, g);
            var i = void 0;
            for (var j = 0; j < h.count; j++) {
                i = h.item(j);
                a.add({ $type: Point_$type, x: d(i), y: e(i) });
            }
        }
    };
    ScatterLineSeries.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case ScatterLineSeries.$$p[0]:
                this.rr(false);
                this.ql();
                break;
        }
    };
    ScatterLineSeries.$t = markType(ScatterLineSeries, 'ScatterLineSeries', ScatterBase.$);
    ScatterLineSeries.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterLineSeries, 'raisePropertyChanged', ['UnknownValuePlotting:acn:acu', [UnknownValuePlotting_$type, enumGetBox(UnknownValuePlotting_$type, 1)]]);
    return ScatterLineSeries;
}(ScatterBase));
export { ScatterLineSeries };
