// igniteui-react-charts/esm5/lib/SeriesPlotAreaMarginHorizontalMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available locations of the axis labels in the chart.
 */
export var SeriesPlotAreaMarginHorizontalMode = /*@__PURE__*/ (function (SeriesPlotAreaMarginHorizontalMode) {
    /**
     * Series horizontal plot area margin mode is determined by the component.
     */
    SeriesPlotAreaMarginHorizontalMode[SeriesPlotAreaMarginHorizontalMode["Auto"] = 0] = "Auto";
    /**
     * The left of the horizontal axis will have a plot area margin, if applicable, and the right of the horizontal axis will have a plot area margin, if applicable.
     */
    SeriesPlotAreaMarginHorizontalMode[SeriesPlotAreaMarginHorizontalMode["LeftMarginRightMargin"] = 1] = "LeftMarginRightMargin";
    /**
     * The left of the horizontal axis will have a range buffer, if applicable, and the right of the horizontal axis will have a range buffer, if applicable.
     */
    SeriesPlotAreaMarginHorizontalMode[SeriesPlotAreaMarginHorizontalMode["LeftBufferRightBuffer"] = 2] = "LeftBufferRightBuffer";
    /**
     * The left of the horizontal axis will have a plot area margin, if applicable, and the right of the horizontal axis will have a range buffer, if applicable.
     */
    SeriesPlotAreaMarginHorizontalMode[SeriesPlotAreaMarginHorizontalMode["LeftMarginRightBuffer"] = 3] = "LeftMarginRightBuffer";
    /**
     * The left of the horizontal axis will have a range buffer, if applicable, and the right of the horizontal axis will have a plot area margin, if applicable.
     */
    SeriesPlotAreaMarginHorizontalMode[SeriesPlotAreaMarginHorizontalMode["LeftBufferRightMargin"] = 4] = "LeftBufferRightMargin";
    /**
     * No horizontal contributions to plot area margins will be made.
     */
    SeriesPlotAreaMarginHorizontalMode[SeriesPlotAreaMarginHorizontalMode["None"] = 5] = "None";
    return SeriesPlotAreaMarginHorizontalMode;
})({});
/**
 * @hidden
 */
export var SeriesPlotAreaMarginHorizontalMode_$type = /*@__PURE__*/ markEnum('SeriesPlotAreaMarginHorizontalMode', 'Auto,0|LeftMarginRightMargin,1|LeftBufferRightBuffer,2|LeftMarginRightBuffer,3|LeftBufferRightMargin,4|None,5');
