// igniteui-react-charts/esm5/lib/SeriesSelectionBehavior.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes the Selection mode to use for the series in the chart.
 */
export var SeriesSelectionBehavior = /*@__PURE__*/ (function (SeriesSelectionBehavior) {
    /**
     * Selection behavior is determined by the component.
     */
    SeriesSelectionBehavior[SeriesSelectionBehavior["Auto"] = 0] = "Auto";
    /**
     * Selection is per data item, multiple data items can be selected.
     */
    SeriesSelectionBehavior[SeriesSelectionBehavior["PerDataItemMultiSelect"] = 1] = "PerDataItemMultiSelect";
    /**
     * Selection is per data item, only one data item can be selected at once.
     */
    SeriesSelectionBehavior[SeriesSelectionBehavior["PerDataItemSingleSelect"] = 2] = "PerDataItemSingleSelect";
    /**
     * Selection is per series and data item, multiple data items can be selected per series at once.
     */
    SeriesSelectionBehavior[SeriesSelectionBehavior["PerSeriesAndDataItemMultiSelect"] = 3] = "PerSeriesAndDataItemMultiSelect";
    /**
     * Selection is per series and data item, only one data item may be selected per series at once.
     */
    SeriesSelectionBehavior[SeriesSelectionBehavior["PerSeriesAndDataItemSingleSelect"] = 4] = "PerSeriesAndDataItemSingleSelect";
    /**
     * Selection is per series and data item, only one item may be selected globally at once.
     */
    SeriesSelectionBehavior[SeriesSelectionBehavior["PerSeriesAndDataItemGlobalSingleSelect"] = 5] = "PerSeriesAndDataItemGlobalSingleSelect";
    /**
     * Selection is per series, multiple entire series may be selected at once.
     */
    SeriesSelectionBehavior[SeriesSelectionBehavior["PerSeriesMultiSelect"] = 6] = "PerSeriesMultiSelect";
    /**
     * Selection is per series, only one series may be selected at once.
     */
    SeriesSelectionBehavior[SeriesSelectionBehavior["PerSeriesSingleSelect"] = 7] = "PerSeriesSingleSelect";
    return SeriesSelectionBehavior;
})({});
/**
 * @hidden
 */
export var SeriesSelectionBehavior_$type = /*@__PURE__*/ markEnum('SeriesSelectionBehavior', 'Auto,0|PerDataItemMultiSelect,1|PerDataItemSingleSelect,2|PerSeriesAndDataItemMultiSelect,3|PerSeriesAndDataItemSingleSelect,4|PerSeriesAndDataItemGlobalSingleSelect,5|PerSeriesMultiSelect,6|PerSeriesSingleSelect,7');
