// igniteui-react-charts/esm5/lib/SeriesPlotAreaMarginVerticalMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available locations of the axis labels in the chart.
 */
export var SeriesPlotAreaMarginVerticalMode = /*@__PURE__*/ (function (SeriesPlotAreaMarginVerticalMode) {
    /**
     * Series horizontal plot area margin mode is determined by the component.
     */
    SeriesPlotAreaMarginVerticalMode[SeriesPlotAreaMarginVerticalMode["Auto"] = 0] = "Auto";
    /**
     * The bottom of the horizontal axis will have a plot area margin, if applicable, and the top of the horizontal axis will have a plot area margin, if applicable.
     */
    SeriesPlotAreaMarginVerticalMode[SeriesPlotAreaMarginVerticalMode["BottomMarginTopMargin"] = 1] = "BottomMarginTopMargin";
    /**
     * The bottom of the horizontal axis will have a range buffer, if applicable, and the top of the horizontal axis will have a range buffer, if applicable.
     */
    SeriesPlotAreaMarginVerticalMode[SeriesPlotAreaMarginVerticalMode["BottomBufferTopBuffer"] = 2] = "BottomBufferTopBuffer";
    /**
     * The bottom of the horizontal axis will have a plot area margin, if applicable, and the top of the horizontal axis will have a range buffer, if applicable.
     */
    SeriesPlotAreaMarginVerticalMode[SeriesPlotAreaMarginVerticalMode["BottomMarginTopBuffer"] = 3] = "BottomMarginTopBuffer";
    /**
     * The bottom of the horizontal axis will have a range buffer, if applicable, and the top of the horizontal axis will have a plot area margin, if applicable.
     */
    SeriesPlotAreaMarginVerticalMode[SeriesPlotAreaMarginVerticalMode["BottomBufferTopMargin"] = 4] = "BottomBufferTopMargin";
    /**
     * No horizontal contributions to plot area margins will be made.
     */
    SeriesPlotAreaMarginVerticalMode[SeriesPlotAreaMarginVerticalMode["None"] = 5] = "None";
    return SeriesPlotAreaMarginVerticalMode;
})({});
/**
 * @hidden
 */
export var SeriesPlotAreaMarginVerticalMode_$type = /*@__PURE__*/ markEnum('SeriesPlotAreaMarginVerticalMode', 'Auto,0|BottomMarginTopMargin,1|BottomBufferTopBuffer,2|BottomMarginTopBuffer,3|BottomBufferTopMargin,4|None,5');
