// igniteui-react-charts/esm5/lib/SeriesViewerScrollbarMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes the scrollbar mode to use for the series viewer.
 */
export var SeriesViewerScrollbarMode = /*@__PURE__*/ (function (SeriesViewerScrollbarMode) {
    /**
     * No horizontal scrollbar will be visible.
     */
    SeriesViewerScrollbarMode[SeriesViewerScrollbarMode["None"] = 0] = "None";
    /**
     * Fading scrollbar will be used that shows only when zoom/pan is changing or the cursor is near.
     */
    SeriesViewerScrollbarMode[SeriesViewerScrollbarMode["Fading"] = 1] = "Fading";
    /**
     * A scrollbar that faides to a hairline will be used that shows only when zoom/pan is changing or the cursor is near.
     */
    SeriesViewerScrollbarMode[SeriesViewerScrollbarMode["FadeToLine"] = 2] = "FadeToLine";
    /**
     * Persistent scrollbar will be used that shows all the time.
     */
    SeriesViewerScrollbarMode[SeriesViewerScrollbarMode["Persistent"] = 3] = "Persistent";
    return SeriesViewerScrollbarMode;
})({});
/**
 * @hidden
 */
export var SeriesViewerScrollbarMode_$type = /*@__PURE__*/ markEnum('SeriesViewerScrollbarMode', 'None,0|Fading,1|FadeToLine,2|Persistent,3');
