// igniteui-react-charts/esm5/lib/SlowStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { SlowStochasticOscillatorIndicatorStrategy } from "./SlowStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
/**
 * @hidden
 */
var SlowStochasticOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SlowStochasticOscillatorIndicator, _super);
    function SlowStochasticOscillatorIndicator() {
        var _this = _super.call(this) || this;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 10;
        _this.trendLinePeriod = 3;
        return _this;
    }
    SlowStochasticOscillatorIndicator.prototype.bj = function () {
        return new SlowStochasticOscillatorIndicator();
    };
    SlowStochasticOscillatorIndicator.prototype.get_abv = function () {
        return new SlowStochasticOscillatorIndicatorStrategy();
    };
    Object.defineProperty(SlowStochasticOscillatorIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    SlowStochasticOscillatorIndicator.prototype.get_ab8 = function () {
        return SlowStochasticOscillatorIndicator.$;
    };
    Object.defineProperty(SlowStochasticOscillatorIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SlowStochasticOscillatorIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(SlowStochasticOscillatorIndicator.ach));
        },
        set: function (a) {
            this.h(SlowStochasticOscillatorIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    SlowStochasticOscillatorIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    SlowStochasticOscillatorIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    SlowStochasticOscillatorIndicator.$t = markType(SlowStochasticOscillatorIndicator, 'SlowStochasticOscillatorIndicator', StrategyBasedIndicator.$);
    SlowStochasticOscillatorIndicator.ach = StrategyBasedIndicator.acd(14, SlowStochasticOscillatorIndicator.$);
    return SlowStochasticOscillatorIndicator;
}(StrategyBasedIndicator));
export { SlowStochasticOscillatorIndicator };
