// igniteui-react-charts/esm5/lib/SortedReadOnlyDoubleCollection.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __generator } from "tslib";
import { Base, IList$1_$type, Number_$type, toEnum, markType } from "igniteui-react-core";
import { ReadOnlyCollection$1 } from "igniteui-react-core";
/**
 * @hidden
 */
var SortedReadOnlyDoubleCollection = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SortedReadOnlyDoubleCollection, _super);
    function SortedReadOnlyDoubleCollection(a, b) {
        var _this = _super.call(this) || this;
        _this.b = null;
        _this.a = null;
        _this.b = new ReadOnlyCollection$1(Number_$type, 1, a);
        _this.a = b;
        return _this;
    }
    SortedReadOnlyDoubleCollection.prototype.indexOf = function (a) {
        var b = this.b.indexOf(a);
        return this.a.indexOf(b);
    };
    SortedReadOnlyDoubleCollection.prototype.insert = function (a, b) {
        this.b.insert(a, b);
    };
    SortedReadOnlyDoubleCollection.prototype.removeAt = function (a) {
        this.b.removeAt(a);
    };
    SortedReadOnlyDoubleCollection.prototype.item = function (a, b) {
        if (arguments.length === 2) {
            this.b.item(a, b);
            return b;
        }
        else {
            var b_1 = this.a.item(a);
            return this.b.item(b_1);
        }
    };
    SortedReadOnlyDoubleCollection.prototype.add = function (a) {
        this.b.add(a);
    };
    SortedReadOnlyDoubleCollection.prototype.clear = function () {
        this.b.clear();
    };
    SortedReadOnlyDoubleCollection.prototype.contains = function (a) {
        return this.b.contains(a);
    };
    SortedReadOnlyDoubleCollection.prototype.copyTo = function (a, b) {
        for (var c = b; c < a.length; c++) {
            a[c] = this.item(c);
        }
    };
    Object.defineProperty(SortedReadOnlyDoubleCollection.prototype, "count", {
        get: function () {
            return this.b.count;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SortedReadOnlyDoubleCollection.prototype, "isReadOnly", {
        get: function () {
            return this.b.isReadOnly;
        },
        enumerable: false,
        configurable: true
    });
    SortedReadOnlyDoubleCollection.prototype.remove = function (a) {
        return this.b.remove(a);
    };
    SortedReadOnlyDoubleCollection.prototype._getEnumerator = function () {
        var a;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    a = 0;
                    _a.label = 1;
                case 1:
                    if (!(a < this.b.count))
                        return [3 /*break*/, 4];
                    return [4 /*yield*/, this.item(a)];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3:
                    a++;
                    return [3 /*break*/, 1];
                case 4: return [2 /*return*/];
            }
        });
    };
    SortedReadOnlyDoubleCollection.prototype.getEnumerator = function () {
        var _this = this;
        return toEnum(function () { return _this._getEnumerator(); }).getEnumerator();
    };
    SortedReadOnlyDoubleCollection.prototype.getEnumeratorObject = function () {
        return this.getEnumerator();
    };
    SortedReadOnlyDoubleCollection.$t = markType(SortedReadOnlyDoubleCollection, 'SortedReadOnlyDoubleCollection', Base.$, [IList$1_$type.specialize(Number_$type)]);
    return SortedReadOnlyDoubleCollection;
}(Base));
export { SortedReadOnlyDoubleCollection };
