// igniteui-react-charts/esm5/lib/SparklineDisplayType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Represents the type of sparkline
 */
export var SparklineDisplayType = /*@__PURE__*/ (function (SparklineDisplayType) {
    /**
     * Display the sparkline as a line.
     */
    SparklineDisplayType[SparklineDisplayType["Line"] = 0] = "Line";
    /**
     * Display the sparkline as a filled polygon.
     */
    SparklineDisplayType[SparklineDisplayType["Area"] = 1] = "Area";
    /**
     * Display the sparkline as a set of columns.
     */
    SparklineDisplayType[SparklineDisplayType["Column"] = 2] = "Column";
    /**
     * Display the sparkline as a set of columns on a boolean scale.
     */
    SparklineDisplayType[SparklineDisplayType["WinLoss"] = 3] = "WinLoss";
    return SparklineDisplayType;
})({});
/**
 * @hidden
 */
export var SparklineDisplayType_$type = /*@__PURE__*/ markEnum('SparklineDisplayType', 'Line,0|Area,1|Column,2|WinLoss,3');
