// igniteui-react-charts/esm5/lib/SparkFrame.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, Array_$type, Point_$type, markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
/**
 * @hidden
 */
var SparkFrame = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(SparkFrame, _super);
    function SparkFrame() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.a = new List$1(Array_$type, 0);
        _this.f = new List$1(Point_$type, 0);
        _this.d = new List$1(Point_$type, 0);
        _this.e = new List$1(Point_$type, 0);
        _this.c = new List$1(Point_$type, 0);
        _this.b = new List$1(Point_$type, 0);
        _this.m = null;
        _this.n = null;
        return _this;
    }
    Object.defineProperty(SparkFrame.prototype, "g", {
        get: function () {
            return this.a;
        },
        set: function (a) {
            this.a = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SparkFrame.prototype, "l", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SparkFrame.prototype, "j", {
        get: function () {
            return this.d;
        },
        set: function (a) {
            this.d = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SparkFrame.prototype, "k", {
        get: function () {
            return this.e;
        },
        set: function (a) {
            this.e = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SparkFrame.prototype, "i", {
        get: function () {
            return this.c;
        },
        set: function (a) {
            this.c = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SparkFrame.prototype, "h", {
        get: function () {
            return this.b;
        },
        set: function (a) {
            this.b = a;
        },
        enumerable: false,
        configurable: true
    });
    SparkFrame.$t = markType(SparkFrame, 'SparkFrame');
    return SparkFrame;
}(Base));
export { SparkFrame };
