// igniteui-react-charts/esm5/lib/Stacked100ColumnSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StackedColumnSeries } from "./StackedColumnSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { Defaults } from "./Defaults";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { isNaN_ } from "igniteui-react-core";
import { StackedColumnSeriesView } from "./StackedColumnSeriesView";
import { Stacked100ColumnBucketCalculator } from "./Stacked100ColumnBucketCalculator";
/**
 * @hidden
 */
var Stacked100ColumnSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100ColumnSeries, _super);
    function Stacked100ColumnSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100ColumnSeries.$;
        return _this;
    }
    Stacked100ColumnSeries.prototype.get_aa4 = function () {
        return true;
    };
    Stacked100ColumnSeries.prototype.jk = function () {
        if (!isNaN_(this.jj)) {
            return _super.prototype.jk.call(this);
        }
        return Defaults.h;
    };
    Stacked100ColumnSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rr(false);
    };
    Stacked100ColumnSeries.prototype.cs = function () {
        return new Stacked100ColumnSeriesView(this);
    };
    Stacked100ColumnSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.ab9 = a;
    };
    Stacked100ColumnSeries.prototype.aav = function () {
        return this.ab9;
    };
    Stacked100ColumnSeries.prototype.abl = function () {
        _super.prototype.abl.call(this);
        Stacked100DataPreparer.a(this);
    };
    Stacked100ColumnSeries.prototype.aba = function (a, b) {
        var c = 0;
        for (var d = 0; d < this.series.count; d++) {
            c += this.series._inner[d].ak.item(b);
        }
        return (this.series._inner[a].ak.item(b) / c) * 100;
    };
    Object.defineProperty(Stacked100ColumnSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zq;
        },
        enumerable: false,
        configurable: true
    });
    Stacked100ColumnSeries.$t = markType(Stacked100ColumnSeries, 'Stacked100ColumnSeries', StackedColumnSeries.$, [IStacked100Series_$type]);
    return Stacked100ColumnSeries;
}(StackedColumnSeries));
export { Stacked100ColumnSeries };
/**
 * @hidden
 */
var Stacked100ColumnSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100ColumnSeriesView, _super);
    function Stacked100ColumnSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.da = null;
        _this.da = a;
        return _this;
    }
    Stacked100ColumnSeriesView.prototype.cx = function () {
        return new Stacked100ColumnBucketCalculator(this);
    };
    Stacked100ColumnSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c5.xa = Defaults.c;
            this.c5.jp = Defaults.i;
        }
    };
    Stacked100ColumnSeriesView.$t = markType(Stacked100ColumnSeriesView, 'Stacked100ColumnSeriesView', StackedColumnSeriesView.$);
    return Stacked100ColumnSeriesView;
}(StackedColumnSeriesView));
export { Stacked100ColumnSeriesView };
