// igniteui-react-charts/esm5/lib/Stacked100BarSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StackedBarSeries } from "./StackedBarSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { Defaults } from "./Defaults";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { isNaN_ } from "igniteui-react-core";
import { StackedBarSeriesView } from "./StackedBarSeriesView";
import { Stacked100BarBucketCalculator } from "./Stacked100BarBucketCalculator";
/**
 * @hidden
 */
var Stacked100BarSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100BarSeries, _super);
    function Stacked100BarSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100BarSeries.$;
        return _this;
    }
    Stacked100BarSeries.prototype.get_aa4 = function () {
        return true;
    };
    Stacked100BarSeries.prototype.jk = function () {
        if (!isNaN_(this.jj)) {
            return _super.prototype.jk.call(this);
        }
        return Defaults.h;
    };
    Stacked100BarSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rr(false);
    };
    Stacked100BarSeries.prototype.cs = function () {
        return new Stacked100BarSeriesView(this);
    };
    Stacked100BarSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.ab9 = a;
    };
    Stacked100BarSeries.prototype.aav = function () {
        return this.ab9;
    };
    Stacked100BarSeries.prototype.abl = function () {
        _super.prototype.abl.call(this);
        Stacked100DataPreparer.a(this);
    };
    Stacked100BarSeries.prototype.aba = function (a, b) {
        var c = 0;
        for (var d = 0; d < this.series.count; d++) {
            c += this.series._inner[d].ak.item(b);
        }
        return (this.series._inner[a].ak.item(b) / c) * 100;
    };
    Object.defineProperty(Stacked100BarSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zq;
        },
        enumerable: false,
        configurable: true
    });
    Stacked100BarSeries.$t = markType(Stacked100BarSeries, 'Stacked100BarSeries', StackedBarSeries.$, [IStacked100Series_$type]);
    return Stacked100BarSeries;
}(StackedBarSeries));
export { Stacked100BarSeries };
/**
 * @hidden
 */
var Stacked100BarSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100BarSeriesView, _super);
    function Stacked100BarSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.db = null;
        _this.db = a;
        return _this;
    }
    Stacked100BarSeriesView.prototype.cx = function () {
        return new Stacked100BarBucketCalculator(this);
    };
    Stacked100BarSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c5.xa = Defaults.c;
            this.c5.jp = Defaults.i;
        }
    };
    Stacked100BarSeriesView.$t = markType(Stacked100BarSeriesView, 'Stacked100BarSeriesView', StackedBarSeriesView.$);
    return Stacked100BarSeriesView;
}(StackedBarSeriesView));
export { Stacked100BarSeriesView };
