// igniteui-react-charts/esm5/lib/Stacked100AreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StackedAreaSeries } from "./StackedAreaSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedAreaSeriesView } from "./StackedAreaSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
var Stacked100AreaSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100AreaSeries, _super);
    function Stacked100AreaSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100AreaSeries.$;
        return _this;
    }
    Stacked100AreaSeries.prototype.get_aa4 = function () {
        return true;
    };
    Stacked100AreaSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rr(false);
    };
    Stacked100AreaSeries.prototype.cs = function () {
        return new Stacked100AreaSeriesView(this);
    };
    Stacked100AreaSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.ab3 = a;
    };
    Stacked100AreaSeries.prototype.abl = function () {
        _super.prototype.abl.call(this);
        Stacked100DataPreparer.a(this);
    };
    Stacked100AreaSeries.prototype.aba = function (a, b) {
        var c = 0;
        for (var d = 0; d < this.series.count; d++) {
            c += this.series._inner[d].ak.item(b);
        }
        return (this.series._inner[a].ak.item(b) / c) * 100;
    };
    Object.defineProperty(Stacked100AreaSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zq;
        },
        enumerable: false,
        configurable: true
    });
    Stacked100AreaSeries.$t = markType(Stacked100AreaSeries, 'Stacked100AreaSeries', StackedAreaSeries.$, [IStacked100Series_$type]);
    return Stacked100AreaSeries;
}(StackedAreaSeries));
export { Stacked100AreaSeries };
/**
 * @hidden
 */
var Stacked100AreaSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100AreaSeriesView, _super);
    function Stacked100AreaSeriesView(a) {
        return _super.call(this, a) || this;
    }
    Stacked100AreaSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c5.xa = Defaults.e;
        }
    };
    Stacked100AreaSeriesView.$t = markType(Stacked100AreaSeriesView, 'Stacked100AreaSeriesView', StackedAreaSeriesView.$);
    return Stacked100AreaSeriesView;
}(StackedAreaSeriesView));
export { Stacked100AreaSeriesView };
