// igniteui-react-charts/esm5/lib/StackedLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { HorizontalStackedSeriesBase } from "./HorizontalStackedSeriesBase";
import { markType } from "igniteui-react-core";
import { AxisRange } from "./AxisRange";
import { ChartSelection } from "./ChartSelection";
import { List$1 } from "igniteui-react-core";
import { StackedSeriesView } from "./StackedSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
var StackedLineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StackedLineSeries, _super);
    function StackedLineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = StackedLineSeries.$;
        return _this;
    }
    StackedLineSeries.prototype.bj = function () {
        return new StackedLineSeries();
    };
    StackedLineSeries.prototype.get_fk = function () {
        return true;
    };
    StackedLineSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rr(false);
    };
    StackedLineSeries.prototype.cs = function () {
        return new StackedLineSeriesView(this);
    };
    StackedLineSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.ab2 = a;
    };
    StackedLineSeries.prototype.ay = function (a) {
        if (this.itemsSource == null) {
            return null;
        }
        if (a == this.xAxis) {
            return new AxisRange(0, this.dn.count - 1);
        }
        if (a == this.yAxis) {
            this.abl();
            return new AxisRange(this.abc, this.abb);
        }
        return null;
    };
    StackedLineSeries.prototype.preferredCategoryMode = function (a) {
        return 0;
    };
    StackedLineSeries.prototype.ak = function (a, b) {
        var c = new List$1(ChartSelection.$, 0);
        this.p4(a, c, b);
        return c.toArray();
    };
    StackedLineSeries.prototype.p4 = function (a, b, c) {
        this.aac(a, b, c, this.yt, 0);
    };
    Object.defineProperty(StackedLineSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zq;
        },
        enumerable: false,
        configurable: true
    });
    StackedLineSeries.$t = markType(StackedLineSeries, 'StackedLineSeries', HorizontalStackedSeriesBase.$);
    return StackedLineSeries;
}(HorizontalStackedSeriesBase));
export { StackedLineSeries };
/**
 * @hidden
 */
var StackedLineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StackedLineSeriesView, _super);
    function StackedLineSeriesView(a) {
        return _super.call(this, a) || this;
    }
    StackedLineSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c5.xa = Defaults.e;
        }
    };
    StackedLineSeriesView.$t = markType(StackedLineSeriesView, 'StackedLineSeriesView', StackedSeriesView.$);
    return StackedLineSeriesView;
}(StackedSeriesView));
export { StackedLineSeriesView };
