// igniteui-react-charts/esm5/lib/Stacked100SplineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StackedSplineSeries } from "./StackedSplineSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedSplineSeriesView } from "./StackedSplineSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
var Stacked100SplineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100SplineSeries, _super);
    function Stacked100SplineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100SplineSeries.$;
        return _this;
    }
    Stacked100SplineSeries.prototype.cs = function () {
        return new Stacked100SplineSeriesView(this);
    };
    Stacked100SplineSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.ab6 = a;
    };
    Stacked100SplineSeries.prototype.get_aa4 = function () {
        return true;
    };
    Stacked100SplineSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rr(false);
    };
    Stacked100SplineSeries.prototype.abl = function () {
        _super.prototype.abl.call(this);
        Stacked100DataPreparer.a(this);
    };
    Stacked100SplineSeries.prototype.aba = function (a, b) {
        var c = 0;
        for (var d = 0; d < this.series.count; d++) {
            c += this.series._inner[d].ak.item(b);
        }
        return (this.series._inner[a].ak.item(b) / c) * 100;
    };
    Object.defineProperty(Stacked100SplineSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zq;
        },
        enumerable: false,
        configurable: true
    });
    Stacked100SplineSeries.$t = markType(Stacked100SplineSeries, 'Stacked100SplineSeries', StackedSplineSeries.$, [IStacked100Series_$type]);
    return Stacked100SplineSeries;
}(StackedSplineSeries));
export { Stacked100SplineSeries };
/**
 * @hidden
 */
var Stacked100SplineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100SplineSeriesView, _super);
    function Stacked100SplineSeriesView(a) {
        return _super.call(this, a) || this;
    }
    Stacked100SplineSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c5.xa = Defaults.e;
        }
    };
    Stacked100SplineSeriesView.$t = markType(Stacked100SplineSeriesView, 'Stacked100SplineSeriesView', StackedSplineSeriesView.$);
    return Stacked100SplineSeriesView;
}(StackedSplineSeriesView));
export { Stacked100SplineSeriesView };
