// igniteui-react-charts/esm5/lib/Stacked100SplineAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StackedSplineAreaSeries } from "./StackedSplineAreaSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-react-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedSplineAreaSeriesView } from "./StackedSplineAreaSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
var Stacked100SplineAreaSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100SplineAreaSeries, _super);
    function Stacked100SplineAreaSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100SplineAreaSeries.$;
        return _this;
    }
    Stacked100SplineAreaSeries.prototype.cs = function () {
        return new Stacked100SplineAreaSeriesView(this);
    };
    Stacked100SplineAreaSeries.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.ab6 = a;
    };
    Stacked100SplineAreaSeries.prototype.get_aa4 = function () {
        return true;
    };
    Stacked100SplineAreaSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rr(false);
    };
    Stacked100SplineAreaSeries.prototype.abl = function () {
        _super.prototype.abl.call(this);
        Stacked100DataPreparer.a(this);
    };
    Stacked100SplineAreaSeries.prototype.aba = function (a, b) {
        var c = 0;
        for (var d = 0; d < this.series.count; d++) {
            c += this.series._inner[d].ak.item(b);
        }
        return (this.series._inner[a].ak.item(b) / c) * 100;
    };
    Object.defineProperty(Stacked100SplineAreaSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zq;
        },
        enumerable: false,
        configurable: true
    });
    Stacked100SplineAreaSeries.$t = markType(Stacked100SplineAreaSeries, 'Stacked100SplineAreaSeries', StackedSplineAreaSeries.$, [IStacked100Series_$type]);
    return Stacked100SplineAreaSeries;
}(StackedSplineAreaSeries));
export { Stacked100SplineAreaSeries };
/**
 * @hidden
 */
var Stacked100SplineAreaSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100SplineAreaSeriesView, _super);
    function Stacked100SplineAreaSeriesView(a) {
        return _super.call(this, a) || this;
    }
    Stacked100SplineAreaSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c5.xa = Defaults.e;
        }
    };
    Stacked100SplineAreaSeriesView.$t = markType(Stacked100SplineAreaSeriesView, 'Stacked100SplineAreaSeriesView', StackedSplineAreaSeriesView.$);
    return Stacked100SplineAreaSeriesView;
}(StackedSplineAreaSeriesView));
export { Stacked100SplineAreaSeriesView };
