// igniteui-react-charts/esm5/lib/StandardDeviationIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { StandardDeviationIndicatorStrategy } from "./StandardDeviationIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var StandardDeviationIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StandardDeviationIndicator, _super);
    function StandardDeviationIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StandardDeviationIndicator.prototype.bj = function () {
        return new StandardDeviationIndicator();
    };
    StandardDeviationIndicator.prototype.get_abv = function () {
        return new StandardDeviationIndicatorStrategy();
    };
    Object.defineProperty(StandardDeviationIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    StandardDeviationIndicator.prototype.get_ab8 = function () {
        return StandardDeviationIndicator.$;
    };
    Object.defineProperty(StandardDeviationIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StandardDeviationIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(StandardDeviationIndicator.ach));
        },
        set: function (a) {
            this.h(StandardDeviationIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    StandardDeviationIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    StandardDeviationIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    StandardDeviationIndicator.$t = markType(StandardDeviationIndicator, 'StandardDeviationIndicator', StrategyBasedIndicator.$);
    StandardDeviationIndicator.ach = StrategyBasedIndicator.acd(20, StandardDeviationIndicator.$);
    return StandardDeviationIndicator;
}(StrategyBasedIndicator));
export { StandardDeviationIndicator };
