// igniteui-react-charts/esm5/lib/StrategyBasedIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { FinancialIndicator } from "./FinancialIndicator";
import { String_$type, Number_$type, typeCast, fromEnum, markType } from "igniteui-react-core";
import { DependencyProperty } from "igniteui-react-core";
import { PropertyMetadata } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { AxisRange } from "./AxisRange";
import { isNaN_ } from "igniteui-react-core";
/**
 * @hidden
 */
var StrategyBasedIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StrategyBasedIndicator, _super);
    function StrategyBasedIndicator() {
        var _this = _super.call(this) || this;
        _this.ab9();
        return _this;
    }
    StrategyBasedIndicator.prototype.ab9 = function () {
        this.abu = this.abv;
        this.ab = this.ab8;
    };
    StrategyBasedIndicator.prototype.ab1 = function () {
        return -2147483648;
    };
    StrategyBasedIndicator.prototype.aca = function (a) {
    };
    StrategyBasedIndicator.prototype.ab3 = function () {
        return -2147483648;
    };
    StrategyBasedIndicator.prototype.ab0 = function () {
        return -2147483648;
    };
    StrategyBasedIndicator.ace = function (a, b, c) {
        var d = DependencyProperty.i(c, Number_$type, b, new PropertyMetadata(2, a, function (e, f) { return (typeCast(StrategyBasedIndicator.$, e)).raisePropertyChanged(c, f.oldValue, f.newValue); }));
        StrategyBasedIndicator.aby.add(c);
        return d;
    };
    StrategyBasedIndicator.acd = function (a, b) {
        return StrategyBasedIndicator.ace(a, b, "Period");
    };
    StrategyBasedIndicator.acb = function (a, b) {
        return StrategyBasedIndicator.ace(a, b, "LongPeriod");
    };
    StrategyBasedIndicator.acf = function (a, b) {
        return StrategyBasedIndicator.ace(a, b, "ShortPeriod");
    };
    StrategyBasedIndicator.prototype.aa7 = function (a, b) {
        var c = this.xj(a, b);
        var d = this.xk(c);
        return this.abu.b(c, d);
    };
    StrategyBasedIndicator.prototype.aa4 = function (a, b) {
        var c = this.xj(a, b);
        if (b == 0) {
            return false;
        }
        if (!this.yf(this.aa7(a, b))) {
            return false;
        }
        var d = this.xk(c);
        if (this.aa0 != null) {
            c.minimumValue = this.aa0.c;
            c.maximumValue = this.aa0.b;
        }
        var e = this.abu.a(c, d);
        for (var f = 0; f < this.aa9 && f < c.indicatorColumn.count; f++) {
            c.indicatorColumn.item(f, NaN);
        }
        if (this.xp != null && this.abx(c)) {
            this.xp.c1();
        }
        return e;
    };
    StrategyBasedIndicator.prototype.abx = function (a) {
        var e_1, _a;
        if (!isNaN_(a.minimumValue) && !isNaN_(a.maximumValue) && a.specifiesRange) {
            var b = this.aa0;
            this.aa0 = new AxisRange(a.minimumValue, a.maximumValue);
            return this.abw(b, this.aa0);
        }
        var c = 1.7976931348623157E+308;
        var d = -1.7976931348623157E+308;
        try {
            for (var _b = __values(fromEnum(a.indicatorColumn)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var e = _c.value;
                if (!isNaN_(e)) {
                    c = Math.min(c, e);
                    d = Math.max(d, e);
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        var f = this.aa0;
        this.aa0 = new AxisRange(c, d);
        return this.abw(f, this.aa0);
    };
    StrategyBasedIndicator.prototype.abw = function (a, b) {
        if (a == null || b == null) {
            return true;
        }
        if (a.c != b.c) {
            return true;
        }
        if (a.b != b.b) {
            return true;
        }
        return false;
    };
    StrategyBasedIndicator.prototype.ab2 = function (a) {
        if (a == -2147483648) {
            return 0;
        }
        if (a > this.indicatorColumn.count && this.indicatorColumn.count > 0) {
            a = this.indicatorColumn.count - 1;
        }
        if (a < 1) {
            return 1;
        }
        return a;
    };
    StrategyBasedIndicator.prototype.xj = function (a, b) {
        var c = _super.prototype.xj.call(this, a, b);
        c.indicatorColumn = this.indicatorColumn;
        c.period = this.ab2(this.ab1());
        c.shortPeriod = this.ab2(this.ab3());
        c.longPeriod = this.ab2(this.ab0());
        c.multiplier = this.abz();
        return c;
    };
    StrategyBasedIndicator.prototype.abz = function () {
        return 1;
    };
    StrategyBasedIndicator.acc = function (a, b) {
        var c = DependencyProperty.i("Multiplier", Number_$type, b, new PropertyMetadata(2, a, function (d, e) { return (typeCast(StrategyBasedIndicator.$, d)).raisePropertyChanged("Multiplier", e.oldValue, e.newValue); }));
        StrategyBasedIndicator.aby.add("Multiplier");
        return c;
    };
    StrategyBasedIndicator.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        if (StrategyBasedIndicator.aby.contains(b)) {
            if (this.xp != null && !this.xp.c1()) {
                this.xn.b0.e(this.id);
                this.aa4(0, this.indicatorColumn.count);
                if (this.aa1.ca) {
                    this.aa1.b8.g$i();
                }
                this.rr(false);
            }
        }
    };
    StrategyBasedIndicator.$t = markType(StrategyBasedIndicator, 'StrategyBasedIndicator', FinancialIndicator.$);
    StrategyBasedIndicator.aby = new List$1(String_$type, 0);
    return StrategyBasedIndicator;
}(FinancialIndicator));
export { StrategyBasedIndicator };
