// igniteui-react-charts/esm5/lib/TimeAxisLabelFormat.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, INotifyPropertyChanged_$type, PropertyChangedEventArgs, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var TimeAxisLabelFormat = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TimeAxisLabelFormat, _super);
    function TimeAxisLabelFormat() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.n = (0);
        _this.a = null;
        _this.d = null;
        _this.c = null;
        _this.b = null;
        _this.propertyChanged = null;
        return _this;
    }
    Object.defineProperty(TimeAxisLabelFormat.prototype, "o", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            var b = this.o;
            if (a != b) {
                this.n = a;
                this.p("Range", b, this.o);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisLabelFormat.prototype, "e", {
        get: function () {
            return this.a;
        },
        set: function (a) {
            var b = this.e;
            if (a != b) {
                this.a = a;
                this.p("Format", b, this.e);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisLabelFormat.prototype, "l", {
        get: function () {
            return this.d;
        },
        set: function (a) {
            var b = this.l;
            if (a != b) {
                this.d = a;
                this.p("RepeatedYearFormat", b, this.l);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisLabelFormat.prototype, "j", {
        get: function () {
            return this.c;
        },
        set: function (a) {
            var b = this.j;
            if (a != b) {
                this.c = a;
                this.p("RepeatedMonthFormat", b, this.j);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TimeAxisLabelFormat.prototype, "h", {
        get: function () {
            return this.b;
        },
        set: function (a) {
            var b = this.h;
            if (a != b) {
                this.b = a;
                this.p("RepeatedDayFormat", b, this.h);
            }
        },
        enumerable: false,
        configurable: true
    });
    TimeAxisLabelFormat.prototype.p = function (a, b, c) {
        this.q(a, b, c);
        if (this.propertyChanged != null) {
            this.propertyChanged(this, new PropertyChangedEventArgs(a));
        }
    };
    TimeAxisLabelFormat.prototype.q = function (a, b, c) {
    };
    TimeAxisLabelFormat.$t = markType(TimeAxisLabelFormat, 'TimeAxisLabelFormat', Base.$, [INotifyPropertyChanged_$type]);
    return TimeAxisLabelFormat;
}(Base));
export { TimeAxisLabelFormat };
