// igniteui-react-charts/esm5/lib/TransitionInSpeedType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available types of speeds used when transitioning in a series.
 */
export var TransitionInSpeedType = /*@__PURE__*/ (function (TransitionInSpeedType) {
    /**
     * A speed type is automatically selected.
     */
    TransitionInSpeedType[TransitionInSpeedType["Auto"] = 0] = "Auto";
    /**
     * All speeds are normal, data points will arrive at the same time.
     */
    TransitionInSpeedType[TransitionInSpeedType["Normal"] = 1] = "Normal";
    /**
     * Data points will arrive later if their value is further from the start point.
     */
    TransitionInSpeedType[TransitionInSpeedType["ValueScaled"] = 2] = "ValueScaled";
    /**
     * Data points will arrive later if their index is further from the axis origin.
     */
    TransitionInSpeedType[TransitionInSpeedType["IndexScaled"] = 3] = "IndexScaled";
    /**
     * Data points will arrive at random times.
     */
    TransitionInSpeedType[TransitionInSpeedType["Random"] = 4] = "Random";
    return TransitionInSpeedType;
})({});
/**
 * @hidden
 */
export var TransitionInSpeedType_$type = /*@__PURE__*/ markEnum('TransitionInSpeedType', 'Auto,0|Normal,1|ValueScaled,2|IndexScaled,3|Random,4');
