// igniteui-react-charts/esm5/lib/TransitionOutSpeedType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available types of speeds used when transitioning in a series.
 */
export var TransitionOutSpeedType = /*@__PURE__*/ (function (TransitionOutSpeedType) {
    /**
     * A speed type is automatically selected.
     */
    TransitionOutSpeedType[TransitionOutSpeedType["Auto"] = 0] = "Auto";
    /**
     * All speeds are normal, data points will arrive at the same time.
     */
    TransitionOutSpeedType[TransitionOutSpeedType["Normal"] = 1] = "Normal";
    /**
     * Data points will arrive later if their value is further from the start point.
     */
    TransitionOutSpeedType[TransitionOutSpeedType["ValueScaled"] = 2] = "ValueScaled";
    /**
     * Data points will arrive later if their index is further from the axis origin.
     */
    TransitionOutSpeedType[TransitionOutSpeedType["IndexScaled"] = 3] = "IndexScaled";
    /**
     * Data points will arrive at random times.
     */
    TransitionOutSpeedType[TransitionOutSpeedType["Random"] = 4] = "Random";
    return TransitionOutSpeedType;
})({});
/**
 * @hidden
 */
export var TransitionOutSpeedType_$type = /*@__PURE__*/ markEnum('TransitionOutSpeedType', 'Auto,0|Normal,1|ValueScaled,2|IndexScaled,3|Random,4');
