// igniteui-react-charts/esm5/lib/TreemapHighlightingMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var TreemapHighlightingMode = /*@__PURE__*/ (function (TreemapHighlightingMode) {
    /**
     * Highlighting mode is determined by the component.
     */
    TreemapHighlightingMode[TreemapHighlightingMode["Auto"] = 0] = "Auto";
    /**
     * No highlighting is defined.
     */
    TreemapHighlightingMode[TreemapHighlightingMode["None"] = 1] = "None";
    /**
     * Treemap nodes will brighten when hovered.
     */
    TreemapHighlightingMode[TreemapHighlightingMode["Brighten"] = 2] = "Brighten";
    /**
     * Treemap nodes will fade if they aren't the hovered node.
     */
    TreemapHighlightingMode[TreemapHighlightingMode["FadeOthers"] = 3] = "FadeOthers";
    return TreemapHighlightingMode;
})({});
/**
 * @hidden
 */
export var TreemapHighlightingMode_$type = /*@__PURE__*/ markEnum('TreemapHighlightingMode', 'Auto,0|None,1|Brighten,2|FadeOthers,3');
