// igniteui-react-charts/esm5/lib/TreemapFillScaleMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var TreemapFillScaleMode = /*@__PURE__*/ (function (TreemapFillScaleMode) {
    TreemapFillScaleMode[TreemapFillScaleMode["Sum"] = 0] = "Sum";
    TreemapFillScaleMode[TreemapFillScaleMode["GlobalSum"] = 1] = "GlobalSum";
    TreemapFillScaleMode[TreemapFillScaleMode["Value"] = 2] = "Value";
    TreemapFillScaleMode[TreemapFillScaleMode["GlobalValue"] = 3] = "GlobalValue";
    return TreemapFillScaleMode;
})({});
/**
 * @hidden
 */
export var TreemapFillScaleMode_$type = /*@__PURE__*/ markEnum('TreemapFillScaleMode', 'Sum,0|GlobalSum,1|Value,2|GlobalValue,3');
