// igniteui-react-charts/esm5/lib/TRIXIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { TRIXIndicatorStrategy } from "./TRIXIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
/**
 * @hidden
 */
var TRIXIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TRIXIndicator, _super);
    function TRIXIndicator() {
        var _this = _super.call(this) || this;
        _this.aa3 = 0;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 9;
        _this.trendLinePeriod = 9;
        return _this;
    }
    TRIXIndicator.prototype.bj = function () {
        return new TRIXIndicator();
    };
    TRIXIndicator.prototype.get_abv = function () {
        return new TRIXIndicatorStrategy();
    };
    Object.defineProperty(TRIXIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    TRIXIndicator.prototype.get_ab8 = function () {
        return TRIXIndicator.$;
    };
    Object.defineProperty(TRIXIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TRIXIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(TRIXIndicator.ach));
        },
        set: function (a) {
            this.h(TRIXIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    TRIXIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    TRIXIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    TRIXIndicator.$t = markType(TRIXIndicator, 'TRIXIndicator', StrategyBasedIndicator.$);
    TRIXIndicator.ach = StrategyBasedIndicator.acd(15, TRIXIndicator.$);
    return TRIXIndicator;
}(StrategyBasedIndicator));
export { TRIXIndicator };
