// igniteui-react-charts/esm5/lib/ValueAxisLabelLocation.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available locations of the y-axis labels in the chart.
 */
export var ValueAxisLabelLocation = /*@__PURE__*/ (function (ValueAxisLabelLocation) {
    /**
     * Places the y-axis labels to the left, outside of the plotting area.
     */
    ValueAxisLabelLocation[ValueAxisLabelLocation["OutsideLeft"] = 0] = "OutsideLeft";
    /**
     * Places the y-axis labels to the right, outside of the plotting area.
     */
    ValueAxisLabelLocation[ValueAxisLabelLocation["OutsideRight"] = 1] = "OutsideRight";
    /**
     * Places the y-axis labels inside the plotting area and to the left of the axis line.
     * CrossingAxis should be set for this setting to take effect.
     */
    ValueAxisLabelLocation[ValueAxisLabelLocation["InsideLeft"] = 2] = "InsideLeft";
    /**
     * Places the y-axis labels inside the plotting area and to the right of the axis line.
     * CrossingAxis should be set for this setting to take effect.
     */
    ValueAxisLabelLocation[ValueAxisLabelLocation["InsideRight"] = 3] = "InsideRight";
    /**
     * Places the y-axis labels based on the type of chart component.
     * Axis Labels Outside Right for Financial Chart and Outside Left for other charts.
     */
    ValueAxisLabelLocation[ValueAxisLabelLocation["Auto"] = 4] = "Auto";
    return ValueAxisLabelLocation;
})({});
/**
 * @hidden
 */
export var ValueAxisLabelLocation_$type = /*@__PURE__*/ markEnum('ValueAxisLabelLocation', 'OutsideLeft,0|OutsideRight,1|InsideLeft,2|InsideRight,3|Auto,4');
