// igniteui-react-charts/esm5/lib/ValueOverlay_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Series } from "./Series";
import { IHasCategoryModePreference_$type } from "./IHasCategoryModePreference";
import { Axis } from "./Axis";
import { DependencyProperty } from "igniteui-react-core";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { Point_$type, typeCast, Delegate_$type, markType, markDep } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
import { Thickness } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
import { Dictionary$2 } from "igniteui-react-core";
import { Pool$1 } from "igniteui-react-core";
import { AxisAnnotation } from "./AxisAnnotation";
import { SeriesView } from "./SeriesView";
import { NotImplementedException } from "igniteui-react-core";
import { AxisRange } from "./AxisRange";
import { NumericAxisBase } from "./NumericAxisBase";
import { AxisAnnotationFrame } from "./AxisAnnotationFrame";
import { ScalerParams } from "./ScalerParams";
import { CategoryXAxis } from "./CategoryXAxis";
import { AxisAnnotationFrameManager } from "./AxisAnnotationFrameManager";
import { PropertyMetadata } from "igniteui-react-core";
import { DeviceUtils } from "igniteui-react-core";
import { isNaN_, isInfinity } from "igniteui-react-core";
import { stringFormat } from "igniteui-react-core";
import { Path } from "igniteui-react-core";
/**
 * @hidden
 */
var ValueOverlay = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ValueOverlay, _super);
    function ValueOverlay() {
        var _this = _super.call(this) || this;
        _this.xf = 0;
        _this.xg = 0;
        _this.xc = 0;
        _this.wz = null;
        _this.w5 = null;
        _this.w4 = null;
        _this.ab = ValueOverlay.$;
        return _this;
    }
    ValueOverlay.prototype.bj = function () {
        return new ValueOverlay();
    };
    ValueOverlay.prototype.i5 = function (a, b, c) {
        return this.xh;
    };
    ValueOverlay.prototype.xd = function (a, b) {
        var c = Math.sqrt(Math.pow(a.x - b.x, 2) + Math.pow(a.y - b.y, 2));
        var d = Math.acos((a.x - b.x) / c);
        if ((a.y - b.y) < 0) {
            d = (2 * Math.PI) - d;
        }
        return d;
    };
    ValueOverlay.prototype.wc = function (a, b, c) {
        if (this.wy == null) {
            return { $type: Point_$type, x: NaN, y: NaN };
        }
        var d = this.xe();
        var e = NaN;
        var f = NaN;
        var g = NaN;
        var h = { $type: Point_$type, x: 0.5, y: 0.5 };
        var i = NaN;
        switch (this.wy.a2) {
            case 2:
                i = Math.sqrt(Math.pow(a.x - h.x, 2) + Math.pow(a.y - h.y, 2));
                g = d;
                e = h.x + Math.cos(g) * i;
                f = h.y + Math.sin(g) * i;
                f = this.cv.bx.top + this.cv.bx.height * (f - this.cv.by.top) / this.cv.by.height;
                e = this.cv.bx.left + this.cv.bx.width * (e - this.cv.by.left) / this.cv.by.width;
                break;
            case 0:
                e = d;
                f = a.y;
                f = this.cv.bx.top + this.cv.bx.height * (f - this.cv.by.top) / this.cv.by.height;
                break;
            case 1:
                f = d;
                e = a.x;
                e = this.cv.bx.left + this.cv.bx.width * (e - this.cv.by.left) / this.cv.by.width;
                break;
            case 3:
                g = this.xd(a, h);
                i = d;
                e = h.x + Math.cos(g) * i;
                f = h.y + Math.sin(g) * i;
                f = this.cv.bx.top + this.cv.bx.height * (f - this.cv.by.top) / this.cv.by.height;
                e = this.cv.bx.left + this.cv.bx.width * (e - this.cv.by.left) / this.cv.by.width;
                break;
        }
        return { $type: Point_$type, x: e, y: f };
    };
    ValueOverlay.prototype.j3 = function (a) {
        throw new NotImplementedException(0);
    };
    ValueOverlay.prototype.kn = function (a) {
        return this;
    };
    ValueOverlay.prototype.gd = function (a) {
        return false;
    };
    ValueOverlay.prototype.ay = function (a) {
        var b = typeCast(NumericAxisBase.$, a);
        var c = this.xh;
        if (isNaN_(c) || isInfinity(c)) {
            return null;
        }
        return new AxisRange(c, c);
    };
    ValueOverlay.prototype.rs = function (a) {
        this.o5(true, this.cv);
        if (!this.w3()) {
            if (this.wz != null) {
                this.x0(NaN, this.d3());
            }
            return;
        }
        if (this.gf(a)) {
            if (this.d3()) {
                if (this.dc.g) {
                    this.dc.t();
                }
                var b = this.xf;
                this.xf = this.xg;
                this.xg = b;
            }
            else {
                this.xf = this.xc;
            }
            this.xc = this.xe();
            this.r7();
        }
        else {
            this.xc = this.xe();
            this.xz(this.xc, this.w1);
        }
    };
    ValueOverlay.prototype.xe = function () {
        var a, b;
        var c = this.getViewInfo(b, a);
        b = c.p0;
        a = c.p1;
        var d = this.getEffectiveViewport1(this.cv);
        var e = new ScalerParams(0, a, b, this.wy.ch, d);
        return this.wy.el(this.xh, e);
    };
    ValueOverlay.prototype.xz = function (a, b) {
        if (this.wy == null) {
            return;
        }
        var c = typeCast(CategoryAxisBase.$, this.wy);
        if (c != null && !c.ce) {
            if (this.xh > c.lk - 1) {
                return;
            }
            if (c.categoryMode != 0) {
                var d = this.getEffectiveViewport1(b);
                var e = 0.5 * c.getCategorySize(b.by, b.bx, d);
                var f = false;
                if (c.cp && !c.ch) {
                    f = true;
                }
                if (typeCast(CategoryXAxis.$, c) !== null && c.ch) {
                    f = true;
                }
                if (f) {
                    e = -e;
                }
                a += e;
            }
        }
        var g = typeCast(NumericAxisBase.$, this.wy);
        if (g != null) {
            if (this.xh < g.l4 || this.xh > g.l3) {
                return;
            }
            if (g.lp && this.xh < 0) {
                return;
            }
        }
        var h;
        switch (this.wy.a2) {
            case 2:
            case 3:
                h = this.wy.j8(a, b.by, b.bx, this.getEffectiveViewport1(b), this.wj(b));
                break;
            default:
                h = this.wy.j8(a, b.by, b.bx, this.getEffectiveViewport1(b), this.wj(b));
                break;
        }
        if (h != null) {
            b.b1(h);
        }
        if (!b.q) {
            this.x0(a, this.d3());
        }
    };
    ValueOverlay.prototype.he = function (a, b, c) {
        var d = true;
        d = _super.prototype.he.call(this, a, b, c);
        if (!c.o() || b.isEmpty || a.isEmpty || this.wy == null || this.wy.bi == null) {
            d = false;
        }
        return d;
    };
    ValueOverlay.prototype.sr = function (a, b) {
        this.rr(false);
    };
    ValueOverlay.prototype.sq = function (a, b) {
        this.rr(false);
    };
    ValueOverlay.prototype.pb = function (a, b, c, d) {
        throw new NotImplementedException(0);
    };
    ValueOverlay.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case ValueOverlay.$$p[11]:
                if (c != null) {
                    c.b0(this);
                }
                if (d != null) {
                    d.ct(this);
                }
                this.rr(false);
                this.ql();
                break;
            case "TransitionProgress":
                this.xg = this.xf + this.jq * (this.xc - this.xf);
                if (this.d5(this.cv)) {
                    return;
                }
                if (this.jq == 1) {
                    this.xz(this.xc, this.w1);
                    this.qo();
                }
                else {
                    this.xz(this.xg, this.w1);
                }
                if (this.db != null) {
                    this.db.qo(this);
                }
                break;
            case ValueOverlay.$$p[13]:
                var e = typeCast(NumericAxisBase.$, this.wy);
                if (e != null) {
                    e.c1();
                }
                this.rr(true);
                this.qm();
                break;
            case ValueOverlay.$$p[12]:
            case ValueOverlay.$$p[10]:
            case ValueOverlay.$$p[1]:
            case ValueOverlay.$$p[0]:
            case ValueOverlay.$$p[3]:
            case ValueOverlay.$$p[4]:
            case ValueOverlay.$$p[9]:
            case ValueOverlay.$$p[5]:
            case ValueOverlay.$$p[6]:
            case ValueOverlay.$$p[7]:
            case ValueOverlay.$$p[8]:
                this.rr(true);
                break;
        }
    };
    ValueOverlay.prototype.qb = function () {
        _super.prototype.qb.call(this);
        if (this.wy != null) {
            this.wy.ib(false);
        }
    };
    ValueOverlay.prototype.o5 = function (a, b) {
        _super.prototype.o5.call(this, a, b);
        var c = b;
        c.b0();
    };
    ValueOverlay.prototype.w3 = function () {
        if (this.wy == null || this.wy.bi == null || isNaN_(this.xh) || this._visibility != 0) {
            return false;
        }
        var a = typeCast(CategoryAxisBase.$, this.wy);
        if (a != null && a.lk == 0) {
            return false;
        }
        return true;
    };
    ValueOverlay.prototype.preferredCategoryMode = function (a) {
        return 0;
    };
    Object.defineProperty(ValueOverlay.prototype, "categoryAxis", {
        get: function () {
            return typeCast(CategoryAxisBase.$, this.wy);
        },
        enumerable: false,
        configurable: true
    });
    ValueOverlay.prototype.cs = function () {
        return new ValueOverlayView(this);
    };
    ValueOverlay.prototype.q1 = function (a) {
        _super.prototype.q1.call(this, a);
        this.w1 = typeCast(ValueOverlayView.$, a);
    };
    ValueOverlay.prototype.rt = function (a, b) {
        _super.prototype.rt.call(this, a, b);
        if (!this.g0) {
            this.cv.a4(b);
            return;
        }
        this.cv.a4(b);
        if (this.d5(this.cu) || !this.w3()) {
            return;
        }
        var c = this.getEffectiveViewport1(this.cv);
        var d = this.wy.el(this.xh, new ScalerParams(0, this.cu.by, a, this.wy.ch, c));
        this.xz(d, typeCast(ValueOverlayView.$, this.cu));
        this.g0 = false;
    };
    ValueOverlay.prototype.getOffsetValue = function () {
        return 0;
    };
    ValueOverlay.prototype.getCategoryWidth = function () {
        return 0;
    };
    Object.defineProperty(ValueOverlay.prototype, "yi", {
        get: function () {
            return new Thickness(1, this.w8, this.xa, this.w9, this.w7);
        },
        enumerable: false,
        configurable: true
    });
    ValueOverlay.prototype.ai = function () {
        return [this.wy];
    };
    ValueOverlay.prototype.x0 = function (a, b) {
        if (isNaN_(a)) {
            if (this.wz != null) {
                AxisAnnotationFrameManager.b(this.wz);
                AxisAnnotationFrameManager.d(this, this.wz, true, this.w4, this.w5, b, this.xi, this.xi, this.w0, this.w0);
            }
            return;
        }
        var c = false;
        if (this.w4 == null) {
            c = true;
            this.w5 = new List$1(Axis.$, 0);
            this.w4 = new Dictionary$2(Axis.$, Pool$1.$.specialize(AxisAnnotation.$), 0);
        }
        this.wz = new AxisAnnotationFrame();
        var d = this.u9;
        if (d == null) {
            d = this.vf;
        }
        var e = this.wy.cp;
        AxisAnnotationFrameManager.c(this.wz, { $type: Point_$type, x: a, y: a }, this, this.db, this.getEffectiveViewport(), d, this.yf, this.w6, this.yh, this.yg, this.yi, this.xb, this.yf, this.w6, this.yh, this.yg, this.yi, this.xb, !e && this.w2, e && this.w2);
        AxisAnnotationFrameManager.d(this, this.wz, c, this.w4, this.w5, b, this.xi, this.xi, this.w0, this.w0);
    };
    ValueOverlay.prototype.get_isValueOverlay = function () {
        return true;
    };
    ValueOverlay.prototype.xv = function (a) {
        if (this.wy != null) {
            return this.wy.gf(a);
        }
        return null;
    };
    Object.defineProperty(ValueOverlay.prototype, "xx", {
        get: function () {
            var a = this.xv(this.xh);
            return a != null ? a : stringFormat("{0:0.##}", this.xh);
        },
        enumerable: false,
        configurable: true
    });
    ValueOverlay.$t = markType(ValueOverlay, 'ValueOverlay', Series.$, [IHasCategoryModePreference_$type]);
    ValueOverlay.$$p = markDep(DependencyProperty, PropertyMetadata, ValueOverlay, 'raisePropertyChanged', ['AxisAnnotationBackgroundCornerRadius:w6:x1', [1, DeviceUtils.g(3)], 'AxisAnnotationBackground:yf:x2', [Brush.$, null], 'AxisAnnotationFormatLabel:w0:x3', [Delegate_$type], 'AxisAnnotationInterpolatedValuePrecision:xi:x4', [1, -1], 'AxisAnnotationOutline:yg:x5', [Brush.$, null], 'AxisAnnotationPaddingBottom:w7:x6', [1, NaN], 'AxisAnnotationPaddingLeft:w8:x7', [1, NaN], 'AxisAnnotationPaddingRight:w9:x8', [1, NaN], 'AxisAnnotationPaddingTop:xa:x9', [1, NaN], 'AxisAnnotationStrokeThickness:xb:ya', [1, NaN], 'AxisAnnotationTextColor:yh:yb', [Brush.$, null], 'Axis:wy:yc', [Axis.$, null], 'IsAxisAnnotationEnabled:w2:yd', [0, false], 'Value:xh:ye', [1, 0]]);
    return ValueOverlay;
}(Series));
export { ValueOverlay };
/**
 * @hidden
 */
var ValueOverlayView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ValueOverlayView, _super);
    function ValueOverlayView(a) {
        var _this = _super.call(this, a) || this;
        _this.b2 = null;
        _this.b2 = new Path();
        return _this;
    }
    ValueOverlayView.prototype.bd = function () {
        _super.prototype.bd.call(this);
        this.b2._stroke = this.e.u9;
        this.b2.ad = this.e.ie;
        this.b2.ai = this.e.v2;
        this.b2.aj = this.e.v3;
    };
    ValueOverlayView.prototype.be = function () {
        _super.prototype.be.call(this);
        var a = this.br();
        this.b2._stroke = a;
        this.b2.ad = this.e.ie + 3;
        this.b2._stroke = a;
        this.b2.ad = this.e.ie + 3;
    };
    ValueOverlayView.prototype.b1 = function (a) {
        this.b2.an = a;
        this.as();
    };
    ValueOverlayView.prototype.b0 = function () {
        this.b2.an = null;
        this.as();
    };
    ValueOverlayView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (a.d) {
            a.t(this.b2);
        }
    };
    ValueOverlayView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        this.e.bc.exportPathData(a, this.b2, "ValueOverlayPath", []);
    };
    ValueOverlayView.$t = markType(ValueOverlayView, 'ValueOverlayView', SeriesView.$);
    return ValueOverlayView;
}(SeriesView));
export { ValueOverlayView };
