// igniteui-react-charts/esm5/lib/VerticalAnchoredCategorySeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnchoredCategorySeries } from "./AnchoredCategorySeries";
import { NumericXAxis } from "./NumericXAxis";
import { DependencyProperty } from "igniteui-react-core";
import { CategoryYAxis } from "./CategoryYAxis";
import { IVerticalAnchoredCategorySeriesInteractionManager_$type } from "./IVerticalAnchoredCategorySeriesInteractionManager";
import { DVContainer } from "igniteui-react-core";
import { runOn, typeCast, EnumUtil, markType, markDep } from "igniteui-react-core";
import { Axis } from "./Axis";
import { ScalerParams } from "./ScalerParams";
import { AxisRange } from "./AxisRange";
import { Rect } from "igniteui-react-core";
import { TrendLineType_$type } from "igniteui-react-core";
import { CategoryTrendLineManagerFactory } from "./CategoryTrendLineManagerFactory";
import { PropertyMetadata } from "igniteui-react-core";
import { truncate, isNaN_, isInfinity } from "igniteui-react-core";
/**
 * @hidden
 */
var VerticalAnchoredCategorySeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(VerticalAnchoredCategorySeries, _super);
    function VerticalAnchoredCategorySeries() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.ace = null;
        return _this;
    }
    VerticalAnchoredCategorySeries.prototype.ck = function () {
        if (this.xAxis != null && this.xAxis.ch) {
            return 4;
        }
        return 3;
    };
    VerticalAnchoredCategorySeries.prototype.cl = function () {
        return 1;
    };
    VerticalAnchoredCategorySeries.prototype.o2 = function () {
        _super.prototype.o2.call(this);
        this.xAxis = null;
        this.yAxis = null;
    };
    Object.defineProperty(VerticalAnchoredCategorySeries.prototype, "acf", {
        get: function () {
            var _this = this;
            if (this.ace == null) {
                this.ace = DVContainer.instance.createInstance(IVerticalAnchoredCategorySeriesInteractionManager_$type, function () { return _this.dr.n; });
            }
            return this.ace;
        },
        enumerable: false,
        configurable: true
    });
    VerticalAnchoredCategorySeries.prototype.get_isVertical = function () {
        return true;
    };
    VerticalAnchoredCategorySeries.prototype.ys = function () {
        return this.yAxis;
    };
    VerticalAnchoredCategorySeries.prototype.y9 = function () {
        return this.xAxis;
    };
    VerticalAnchoredCategorySeries.prototype.getOffsetValue = function () {
        return this.yz.r(this.ys(), this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
    };
    VerticalAnchoredCategorySeries.prototype.getCategoryWidth = function () {
        return this.yAxis.getCategorySize(this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv));
    };
    VerticalAnchoredCategorySeries.prototype.j8 = function (a, b) {
        return this.j9(a, b, this.ys(), runOn(this, this.zt), this.valueColumn);
    };
    VerticalAnchoredCategorySeries.prototype.ka = function (a, b) {
        return this.kb(a, b, this.ys(), runOn(this, this.zt), this.valueColumn);
    };
    VerticalAnchoredCategorySeries.prototype.ip = function (a, b, c, d, e) {
        if (this.valueColumn == null) {
            return Number.POSITIVE_INFINITY;
        }
        return this.iq(a, b, this.yAxis, d, e, this.valueColumn.count, runOn(this, this.zt));
    };
    VerticalAnchoredCategorySeries.prototype.az = function (a) {
        if (a.isEmpty && this.db != null) {
            a = this.db.v4;
        }
        if (this.valueColumn == null) {
            return null;
        }
        var b = this.wj(this.cv);
        var c = this.ys();
        var d = new ScalerParams(0, a, this.db.wk, c.ch, this.getEffectiveViewport());
        var e = (c.eo(b.top, d));
        var f = (c.eo(b.bottom, d));
        if (!c.ch) {
            var g = e;
            e = f;
            f = g;
        }
        e -= this.abb;
        f += this.abb;
        var h = null;
        if (c.cn) {
            h = c;
            e = h.i$f(e);
            f = h.i$f(f);
        }
        else {
            e = Math.floor(e);
            f = Math.ceil(f);
        }
        if (isNaN_(e) || isNaN_(f)) {
            return null;
        }
        e = Math.max(0, Math.min(this.valueColumn.count - 1, e));
        f = Math.max(0, Math.min(this.valueColumn.count - 1, f));
        var i = 1.7976931348623157E+308;
        var j = -1.7976931348623157E+308;
        if (truncate(e) <= truncate(f)) {
            for (var k = truncate(e); k <= truncate(f); k++) {
                var l = k;
                if (h != null) {
                    l = h.i$b._inner[k];
                }
                var m = this.valueColumn.item(l);
                if (this.valueColumn.mayContainUnknowns) {
                    if (isInfinity(m) || isNaN_(m)) {
                        continue;
                    }
                }
                i = Math.min(i, m);
                j = Math.max(j, m);
            }
        }
        var n = new AxisRange(i, j);
        n = this.ax(n, this.y9(), a);
        return n;
    };
    VerticalAnchoredCategorySeries.prototype.b2 = function () {
        if (this.b3 == 0) {
            return 1;
        }
        else {
            return _super.prototype.b2.call(this);
        }
    };
    VerticalAnchoredCategorySeries.prototype.wp = function (a) {
        if (!this.hasMarkers || !this.xg()) {
            return Rect.empty;
        }
        return this.acf.e$b(this, this.yu.n, a, runOn(this, this.acn));
    };
    VerticalAnchoredCategorySeries.prototype.acn = function (a) {
        return this.aav.de(a);
    };
    VerticalAnchoredCategorySeries.prototype.gw = function (a, b) {
        if (this.gv(a, b)) {
            return true;
        }
        if (this.xh(a, b)) {
            return true;
        }
        return false;
    };
    VerticalAnchoredCategorySeries.prototype.i5 = function (a, b, c) {
        if (this.db == null) {
            return NaN;
        }
        var d = this.getEffectiveViewport1(this.cv);
        var e = this.ys();
        var f = new ScalerParams(0, this.db.v4, this.cv.bx, e.ch, d);
        var g = this.yz.r(e, this.db.v4, this.cv.bx, d);
        return this.i7(this.valueColumn, a, e, f, g, runOn(this, this.zt), b, c);
    };
    VerticalAnchoredCategorySeries.prototype.wc = function (a, b, c) {
        var d = this.ys();
        return this.we(a, b, c, this.yz.r(d, this.cv.by, this.cv.bx, this.getEffectiveViewport1(this.cv)), this.y9(), d, null, null, null);
    };
    VerticalAnchoredCategorySeries.prototype.fetchXAxis = function () {
        return this.xAxis;
    };
    VerticalAnchoredCategorySeries.prototype.fetchYAxis = function () {
        return this.yAxis;
    };
    VerticalAnchoredCategorySeries.prototype.abr = function (a) {
        this.xAxis = typeCast(NumericXAxis.$, a);
    };
    VerticalAnchoredCategorySeries.prototype.abs = function (a) {
        this.yAxis = typeCast(CategoryYAxis.$, a);
    };
    VerticalAnchoredCategorySeries.prototype.zp = function () {
        return this.xAxis != null && this.xAxis.c1();
    };
    VerticalAnchoredCategorySeries.prototype.ym = function (a, b) {
        return this.aas(a, b, this.yAxis);
    };
    VerticalAnchoredCategorySeries.prototype.ir = function (a) {
        return this.acf.e$a(this, a);
    };
    VerticalAnchoredCategorySeries.prototype.zt = function (a) {
        return this.it(a, this.ys());
    };
    VerticalAnchoredCategorySeries.prototype.j3 = function (a) {
        return truncate(Math.round(this.ir(a)));
    };
    VerticalAnchoredCategorySeries.prototype.kn = function (a) {
        var b = this.j3(a);
        return b >= 0 && this.dn != null && b < this.dn.count ? this.dn.item(b) : null;
    };
    VerticalAnchoredCategorySeries.prototype.q9 = function (a, b, c, d) {
        _super.prototype.q9.call(this, a, b, c, d);
        switch (b) {
            case VerticalAnchoredCategorySeries.$$p[0]:
                if (c != d) {
                    this.pc(typeCast(Axis.$, c));
                    this.rh(typeCast(Axis.$, d));
                    this.y3.cw.g(this.id);
                    if (this.db != null) {
                        this.db.n2();
                    }
                    this.rr(false);
                    this.ql();
                }
                break;
            case VerticalAnchoredCategorySeries.$$p[1]:
                if (c != d) {
                    this.pc(typeCast(Axis.$, c));
                    this.rh(typeCast(Axis.$, d));
                    this.y3.cw.g(this.id);
                    this.zp();
                    this.rr(false);
                    this.ql();
                }
                break;
            case "TrendLineType":
                var e = EnumUtil.getEnumValue(TrendLineType_$type, d);
                if (e != 0) {
                    this.aav.c7 = CategoryTrendLineManagerFactory.a(this.aav.c7, this.yAxis, this.sv, this, this.dr);
                }
                break;
        }
    };
    VerticalAnchoredCategorySeries.prototype.aci = function (a) {
        if (typeCast(CategoryYAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    VerticalAnchoredCategorySeries.prototype.ach = function (a) {
        if (typeCast(NumericXAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    Object.defineProperty(VerticalAnchoredCategorySeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zq;
        },
        enumerable: false,
        configurable: true
    });
    VerticalAnchoredCategorySeries.$t = markType(VerticalAnchoredCategorySeries, 'VerticalAnchoredCategorySeries', AnchoredCategorySeries.$);
    VerticalAnchoredCategorySeries.$$p = markDep(DependencyProperty, PropertyMetadata, VerticalAnchoredCategorySeries, 'raisePropertyChanged', ['XAxis::acl', [NumericXAxis.$, null], 'YAxis::acm', [CategoryYAxis.$, null]]);
    return VerticalAnchoredCategorySeries;
}(AnchoredCategorySeries));
export { VerticalAnchoredCategorySeries };
