// igniteui-react-charts/esm5/lib/ViewportChangedMessage_Spark.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { InteractionMessage_Spark } from "./InteractionMessage_Spark";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
var ViewportChangedMessage_Spark = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ViewportChangedMessage_Spark, _super);
    function ViewportChangedMessage_Spark() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.d = 0;
        _this.c = 0;
        _this.e = 0;
        _this.b = 0;
        return _this;
    }
    Object.defineProperty(ViewportChangedMessage_Spark.prototype, "newTop", {
        get: function () {
            return this.d;
        },
        set: function (a) {
            this.d = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ViewportChangedMessage_Spark.prototype, "newLeft", {
        get: function () {
            return this.c;
        },
        set: function (a) {
            this.c = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ViewportChangedMessage_Spark.prototype, "newWidth", {
        get: function () {
            return this.e;
        },
        set: function (a) {
            this.e = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ViewportChangedMessage_Spark.prototype, "newHeight", {
        get: function () {
            return this.b;
        },
        set: function (a) {
            this.b = a;
        },
        enumerable: false,
        configurable: true
    });
    ViewportChangedMessage_Spark.prototype.toString = function () {
        return "ViewportChangedMessage[" + this.newTop.toString() + ", " + this.newLeft.toString() + ", " + this.newWidth.toString() + ", " + this.newHeight.toString() + "]";
    };
    ViewportChangedMessage_Spark.$t = markType(ViewportChangedMessage_Spark, 'ViewportChangedMessage_Spark', InteractionMessage_Spark.$);
    return ViewportChangedMessage_Spark;
}(InteractionMessage_Spark));
export { ViewportChangedMessage_Spark };
