// igniteui-react-charts/esm5/lib/WilliamsPercentRIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { WilliamsPercentRIndicatorStrategy } from "./WilliamsPercentRIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var WilliamsPercentRIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WilliamsPercentRIndicator, _super);
    function WilliamsPercentRIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    WilliamsPercentRIndicator.prototype.bj = function () {
        return new WilliamsPercentRIndicator();
    };
    WilliamsPercentRIndicator.prototype.get_abv = function () {
        return new WilliamsPercentRIndicatorStrategy();
    };
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "abv", {
        get: function () {
            return this.get_abv();
        },
        enumerable: false,
        configurable: true
    });
    WilliamsPercentRIndicator.prototype.get_ab8 = function () {
        return WilliamsPercentRIndicator.$;
    };
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "ab8", {
        get: function () {
            return this.get_ab8();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "acg", {
        get: function () {
            return typeGetValue(this.c(WilliamsPercentRIndicator.ach));
        },
        set: function (a) {
            this.h(WilliamsPercentRIndicator.ach, a);
        },
        enumerable: false,
        configurable: true
    });
    WilliamsPercentRIndicator.prototype.ab1 = function () {
        return this.acg;
    };
    WilliamsPercentRIndicator.prototype.aca = function (a) {
        this.acg = a;
    };
    WilliamsPercentRIndicator.$t = markType(WilliamsPercentRIndicator, 'WilliamsPercentRIndicator', StrategyBasedIndicator.$);
    WilliamsPercentRIndicator.ach = StrategyBasedIndicator.acd(14, WilliamsPercentRIndicator.$);
    return WilliamsPercentRIndicator;
}(StrategyBasedIndicator));
export { WilliamsPercentRIndicator };
