// igniteui-react-charts/esm5/lib/XAxisLabelLocation.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Describes available locations of the x-axis labels in the chart.
 */
export var XAxisLabelLocation = /*@__PURE__*/ (function (XAxisLabelLocation) {
    /**
     * Places the x-axis labels at the top, outside of the plotting area.
     */
    XAxisLabelLocation[XAxisLabelLocation["OutsideTop"] = 0] = "OutsideTop";
    /**
     * Places the x-axis labels at the bottom, outside of the plotting area
     */
    XAxisLabelLocation[XAxisLabelLocation["OutsideBottom"] = 1] = "OutsideBottom";
    /**
     * Places the x-axis labels inside the plotting area above the axis line.
     * CrossingAxis should be set for this setting to take effect.
     */
    XAxisLabelLocation[XAxisLabelLocation["InsideTop"] = 2] = "InsideTop";
    /**
     * Places the x-axis labels inside the plotting area below the axis line.
     * CrossingAxis should be set for this setting to take effect.
     */
    XAxisLabelLocation[XAxisLabelLocation["InsideBottom"] = 3] = "InsideBottom";
    /**
     * Places the x-axis labels based on the type of chart component.
     * Axis Labels Outside Bottom for all types of charts.
     */
    XAxisLabelLocation[XAxisLabelLocation["Auto"] = 4] = "Auto";
    return XAxisLabelLocation;
})({});
/**
 * @hidden
 */
export var XAxisLabelLocation_$type = /*@__PURE__*/ markEnum('XAxisLabelLocation', 'OutsideTop,0|OutsideBottom,1|InsideTop,2|InsideBottom,3|Auto,4');
