// igniteui-react-charts/esm5/lib/XYChart.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { DomainChart } from "./DomainChart";
import { fromEnum, enumGetBox, markType } from "igniteui-react-core";
import { HorizontalAlignment_$type } from "igniteui-react-core";
import { VerticalAlignment_$type } from "igniteui-react-core";
import { Visibility_$type } from "igniteui-react-core";
import { YAxisLabelLocation_$type } from "./YAxisLabelLocation";
import { XAxisLabelLocation_$type } from "./XAxisLabelLocation";
import { AxisLabelSettings } from "./AxisLabelSettings";
import { AxisDefaults } from "./AxisDefaults";
import { TitleSettings } from "./TitleSettings";
import { ScalerParams } from "./ScalerParams";
import { DeviceUtils } from "igniteui-react-core";
/**
 * @hidden
 */
var XYChart = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(XYChart, _super);
    function XYChart() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.z4 = null;
        _this.z5 = null;
        _this.aal = 0;
        _this.aan = 5;
        _this.aam = 0;
        _this.aak = 5;
        _this.aa2 = 5;
        _this.aa4 = 0;
        _this.aa3 = 5;
        _this.aa1 = 0;
        _this.afv = null;
        _this.af2 = null;
        _this.aft = null;
        _this.afu = null;
        _this.aaw = 0;
        _this.abd = 0;
        _this.aav = NaN;
        _this.abc = NaN;
        _this.aay = NaN;
        _this.abf = NaN;
        _this.aax = NaN;
        _this.abe = NaN;
        _this.aau = NaN;
        _this.abb = NaN;
        _this.af1 = null;
        _this.af8 = null;
        _this.acn = null;
        _this.acr = null;
        _this.acp = null;
        _this.act = null;
        _this.aci = null;
        _this.acj = null;
        _this.afw = null;
        _this.af3 = null;
        _this.aao = DeviceUtils.g(1);
        _this.aa5 = DeviceUtils.g(1);
        _this.aap = DeviceUtils.g(1);
        _this.aa6 = DeviceUtils.g(1);
        _this.afy = null;
        _this.af5 = null;
        _this.afz = null;
        _this.af6 = null;
        _this.aaq = DeviceUtils.g(1);
        _this.aa7 = DeviceUtils.g(1);
        _this.aar = DeviceUtils.g(8);
        _this.aa8 = 0;
        _this.af0 = null;
        _this.af7 = null;
        _this.aas = DeviceUtils.g(1);
        _this.aa9 = DeviceUtils.g(1);
        _this.aco = null;
        _this.acs = null;
        _this.afx = null;
        _this.af4 = null;
        _this.aaj = 0;
        _this.aa0 = 0;
        _this.aai = NaN;
        _this.aaz = NaN;
        _this.aat = 0;
        _this.aba = -90;
        _this.aae = false;
        _this.aaf = false;
        _this.afo = 1;
        _this.agr = 1;
        _this.afn = 1;
        _this.afp = 2;
        _this.agp = 0;
        _this.agq = 1;
        _this.agv = 0;
        _this.agw = 0;
        _this.aac = 4;
        _this.aaa = 4;
        _this.acm = null;
        _this.zy = null;
        _this.acq = null;
        _this.zz = null;
        return _this;
    }
    XYChart.prototype.provideContainer = function (a) {
        if (a != null) {
            var b = this.z2();
            var c = this.z3();
            if (b != null) {
                b.provideRenderer(a);
            }
            if (c != null) {
                c.provideRenderer(a);
            }
        }
        _super.prototype.provideContainer.call(this, a);
    };
    XYChart.prototype.u5 = function () {
        _super.prototype.u5.call(this);
        var a = this.z2();
        var b = this.z3();
        if (a != null && a.az != null) {
            this.agb = a.az.textColor;
        }
        if (b != null && b.az != null) {
            this.agi = b.az.textColor;
        }
        this.agg = AxisDefaults.j;
        this.agn = AxisDefaults.j;
        this.agc = AxisDefaults.h;
        this.agj = AxisDefaults.h;
        this.agf = AxisDefaults.g;
        this.agm = AxisDefaults.g;
    };
    XYChart.prototype.vx = function (a, b, c) {
        var _this = this;
        _super.prototype.vx.call(this, a, b, c);
        switch (a) {
            case "XAxisLabelLeftMargin":
                this.aff(function (d) { return d.leftMargin = _this.abn; });
                break;
            case "XAxisLabelTopMargin":
                this.aff(function (d) { return d.topMargin = _this.abp; });
                break;
            case "XAxisLabelRightMargin":
                this.aff(function (d) { return d.rightMargin = _this.abo; });
                break;
            case "XAxisLabelBottomMargin":
                this.aff(function (d) { return d.bottomMargin = _this.abm; });
                break;
            case "YAxisLabelLeftMargin":
                this.afk(function (d) { return d.leftMargin = _this.ab4; });
                break;
            case "YAxisLabelTopMargin":
                this.afk(function (d) { return d.topMargin = _this.ab6; });
                break;
            case "YAxisLabelRightMargin":
                this.afk(function (d) { return d.rightMargin = _this.ab5; });
                break;
            case "YAxisLabelBottomMargin":
                this.afk(function (d) { return d.bottomMargin = _this.ab3; });
                break;
            case "ActualItemsSource":
            case "ShouldDisplayMockData":
            case "ShouldUseSkeletonStyleForMockData":
                this.af9 = this.yw(this.agb);
                this.aga = this.yw(this.agi);
                break;
            case "XAxisLabelTextColor":
                this.af9 = this.yw(this.agb);
                break;
            case "YAxisLabelTextColor":
                this.aga = this.yw(this.agi);
                break;
            case "ActualXAxisLabelTextColor":
                this.aff(function (d) { return d.textColor = _this.af9; });
                break;
            case "ActualYAxisLabelTextColor":
                this.afk(function (d) { return d.textColor = _this.aga; });
                break;
            case "XAxisTitleMargin":
                this.afh(function (d) { return d.margin = _this.aby; });
                break;
            case "YAxisTitleMargin":
                this.afm(function (d) { return d.margin = _this.acf; });
                break;
            case "XAxisTitleLeftMargin":
                this.afh(function (d) { return d.leftMargin = _this.abx; });
                break;
            case "YAxisTitleLeftMargin":
                this.afm(function (d) { return d.leftMargin = _this.ace; });
                break;
            case "XAxisTitleTopMargin":
                this.afh(function (d) { return d.topMargin = _this.ab0; });
                break;
            case "YAxisTitleTopMargin":
                this.afm(function (d) { return d.topMargin = _this.ach; });
                break;
            case "XAxisTitleRightMargin":
                this.afh(function (d) { return d.rightMargin = _this.abz; });
                break;
            case "YAxisTitleRightMargin":
                this.afm(function (d) { return d.rightMargin = _this.acg; });
                break;
            case "XAxisTitleBottomMargin":
                this.afh(function (d) { return d.bottomMargin = _this.abw; });
                break;
            case "YAxisTitleBottomMargin":
                this.afm(function (d) { return d.bottomMargin = _this.acd; });
                break;
            case "XAxisTitleTextColor":
                this.afh(function (d) { return d.textColor = _this.agh; });
                break;
            case "YAxisTitleTextColor":
                this.afm(function (d) { return d.textColor = _this.ago; });
                break;
            case "XAxisFormatLabel":
                this.afd(function (d) { return d.au = _this.z6; });
                break;
            case "YAxisFormatLabel":
                this.afi(function (d) { return d.au = _this.z7; });
                break;
        }
    };
    Object.defineProperty(XYChart.prototype, "z6", {
        get: function () {
            return this.z4;
        },
        set: function (a) {
            var b = this.z6;
            if (a != b) {
                this.z4 = a;
                this.vu("XAxisFormatLabel", b, this.z6);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "z7", {
        get: function () {
            return this.z5;
        },
        set: function (a) {
            var b = this.z7;
            if (a != b) {
                this.z5 = a;
                this.vu("YAxisFormatLabel", b, this.z7);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abn", {
        get: function () {
            return this.aal;
        },
        set: function (a) {
            var b = this.abn;
            if (a != b) {
                this.aal = a;
                this.vu("XAxisLabelLeftMargin", b, this.abn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abp", {
        get: function () {
            return this.aan;
        },
        set: function (a) {
            var b = this.abp;
            if (a != b) {
                this.aan = a;
                this.vu("XAxisLabelTopMargin", b, this.abp);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abo", {
        get: function () {
            return this.aam;
        },
        set: function (a) {
            var b = this.abo;
            if (a != b) {
                this.aam = a;
                this.vu("XAxisLabelRightMargin", b, this.abo);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abm", {
        get: function () {
            return this.aak;
        },
        set: function (a) {
            var b = this.abm;
            if (a != b) {
                this.aak = a;
                this.vu("XAxisLabelBottomMargin", b, this.abm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab4", {
        get: function () {
            return this.aa2;
        },
        set: function (a) {
            var b = this.ab4;
            if (a != b) {
                this.aa2 = a;
                this.vu("YAxisLabelLeftMargin", b, this.ab4);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab6", {
        get: function () {
            return this.aa4;
        },
        set: function (a) {
            var b = this.ab6;
            if (a != b) {
                this.aa4 = a;
                this.vu("YAxisLabelTopMargin", b, this.ab6);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab5", {
        get: function () {
            return this.aa3;
        },
        set: function (a) {
            var b = this.ab5;
            if (a != b) {
                this.aa3 = a;
                this.vu("YAxisLabelRightMargin", b, this.ab5);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab3", {
        get: function () {
            return this.aa1;
        },
        set: function (a) {
            var b = this.ab3;
            if (a != b) {
                this.aa1 = a;
                this.vu("YAxisLabelBottomMargin", b, this.ab3);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agb", {
        get: function () {
            return this.afv;
        },
        set: function (a) {
            var b = this.agb;
            if (a != b) {
                this.afv = a;
                this.vu("XAxisLabelTextColor", b, this.agb);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agi", {
        get: function () {
            return this.af2;
        },
        set: function (a) {
            var b = this.agi;
            if (a != b) {
                this.af2 = a;
                this.vu("YAxisLabelTextColor", b, this.agi);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "af9", {
        get: function () {
            return this.aft;
        },
        set: function (a) {
            var b = this.aft;
            if (a != b) {
                this.aft = a;
                this.vu("ActualXAxisLabelTextColor", b, this.af9);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aga", {
        get: function () {
            return this.afu;
        },
        set: function (a) {
            var b = this.afu;
            if (a != b) {
                this.afu = a;
                this.vu("ActualYAxisLabelTextColor", b, this.aga);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aby", {
        get: function () {
            return this.aaw;
        },
        set: function (a) {
            var b = this.aby;
            if (a != b) {
                this.aaw = a;
                this.vu("XAxisTitleMargin", b, this.aby);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acf", {
        get: function () {
            return this.abd;
        },
        set: function (a) {
            var b = this.acf;
            if (a != b) {
                this.abd = a;
                this.vu("YAxisTitleMargin", b, this.acf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abx", {
        get: function () {
            return this.aav;
        },
        set: function (a) {
            var b = this.abx;
            if (a != b) {
                this.aav = a;
                this.vu("XAxisTitleLeftMargin", b, this.abx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ace", {
        get: function () {
            return this.abc;
        },
        set: function (a) {
            var b = this.ace;
            if (a != b) {
                this.abc = a;
                this.vu("YAxisTitleLeftMargin", b, this.ace);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab0", {
        get: function () {
            return this.aay;
        },
        set: function (a) {
            var b = this.ab0;
            if (a != b) {
                this.aay = a;
                this.vu("XAxisTitleTopMargin", b, this.ab0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ach", {
        get: function () {
            return this.abf;
        },
        set: function (a) {
            var b = this.ach;
            if (a != b) {
                this.abf = a;
                this.vu("YAxisTitleTopMargin", b, this.ach);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abz", {
        get: function () {
            return this.aax;
        },
        set: function (a) {
            var b = this.abz;
            if (a != b) {
                this.aax = a;
                this.vu("XAxisTitleRightMargin", b, this.abz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acg", {
        get: function () {
            return this.abe;
        },
        set: function (a) {
            var b = this.acg;
            if (a != b) {
                this.abe = a;
                this.vu("YAxisTitleRightMargin", b, this.acg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abw", {
        get: function () {
            return this.aau;
        },
        set: function (a) {
            var b = this.abw;
            if (a != b) {
                this.aau = a;
                this.vu("XAxisTitleBottomMargin", b, this.abw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acd", {
        get: function () {
            return this.abb;
        },
        set: function (a) {
            var b = this.acd;
            if (a != b) {
                this.abb = a;
                this.vu("YAxisTitleBottomMargin", b, this.acd);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agh", {
        get: function () {
            return this.af1;
        },
        set: function (a) {
            var b = this.agh;
            if (a != b) {
                this.af1 = a;
                this.vu("XAxisTitleTextColor", b, this.agh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ago", {
        get: function () {
            return this.af8;
        },
        set: function (a) {
            var b = this.ago;
            if (a != b) {
                this.af8 = a;
                this.vu("YAxisTitleTextColor", b, this.ago);
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.vy = function (a, b, c) {
        var _this = this;
        _super.prototype.vy.call(this, a, b, c);
        switch (a) {
            case "XAxisLabelTextStyle":
                this.aff(function (d) { return d.textStyle = _this.adb; });
                break;
            case "YAxisLabelTextStyle":
                this.afk(function (d) { return d.textStyle = _this.aeg; });
                break;
            case "XAxisTitleTextStyle":
                this.afh(function (d) { return d.textStyle = _this.adz; });
                break;
            case "YAxisTitleTextStyle":
                this.afm(function (d) { return d.textStyle = _this.ae4; });
                break;
        }
    };
    Object.defineProperty(XYChart.prototype, "adb", {
        get: function () {
            return this.acn;
        },
        set: function (a) {
            var b = this.adb;
            if (a != b) {
                this.acn = a;
                this.vu("XAxisLabelTextStyle", b, this.adb);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aeg", {
        get: function () {
            return this.acr;
        },
        set: function (a) {
            var b = this.aeg;
            if (a != b) {
                this.acr = a;
                this.vu("YAxisLabelTextStyle", b, this.aeg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "adz", {
        get: function () {
            return this.acp;
        },
        set: function (a) {
            var b = this.adz;
            if (a != b) {
                this.acp = a;
                this.vu("XAxisTitleTextStyle", b, this.adz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ae4", {
        get: function () {
            return this.act;
        },
        set: function (a) {
            var b = this.ae4;
            if (a != b) {
                this.act = a;
                this.vu("YAxisTitleTextStyle", b, this.ae4);
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.afa = function (a) {
        var _this = this;
        this.ae7(a, function (b) { return b.leftMargin = _this.abn; });
        this.ae7(a, function (b) { return b.topMargin = _this.abp; });
        this.ae7(a, function (b) { return b.rightMargin = _this.abo; });
        this.ae7(a, function (b) { return b.bottomMargin = _this.abm; });
        this.ae7(a, function (b) { return b.textColor = _this.af9; });
        this.ae8(a, function (b) { return b.textColor = _this.agh; });
        this.ae8(a, function (b) { return b.leftMargin = _this.abx; });
        this.ae8(a, function (b) { return b.topMargin = _this.ab0; });
        this.ae8(a, function (b) { return b.rightMargin = _this.abz; });
        this.ae8(a, function (b) { return b.bottomMargin = _this.abw; });
        this.ae7(a, function (b) { return b.textStyle = _this.adb; });
        this.ae8(a, function (b) { return b.textStyle = _this.adz; });
    };
    XYChart.prototype.afc = function (a) {
        var _this = this;
        this.ae7(a, function (b) { return b.leftMargin = _this.ab4; });
        this.ae7(a, function (b) { return b.topMargin = _this.ab6; });
        this.ae7(a, function (b) { return b.rightMargin = _this.ab5; });
        this.ae7(a, function (b) { return b.bottomMargin = _this.ab3; });
        this.ae7(a, function (b) { return b.textColor = _this.aga; });
        this.ae8(a, function (b) { return b.textColor = _this.ago; });
        this.ae8(a, function (b) { return b.leftMargin = _this.ace; });
        this.ae8(a, function (b) { return b.topMargin = _this.ach; });
        this.ae8(a, function (b) { return b.rightMargin = _this.acg; });
        this.ae8(a, function (b) { return b.bottomMargin = _this.acd; });
        this.ae7(a, function (b) { return b.textStyle = _this.aeg; });
        this.ae8(a, function (b) { return b.textStyle = _this.ae4; });
    };
    XYChart.prototype.wo = function (a) {
        var b = this.be(a);
        var c = this.dc(a, b);
        if (c == null) {
            return;
        }
        c.itemsSource = this.h1(a);
        if (a.highlightedData != null) {
            c.highlightedItemsSource = this.h2(a);
        }
        c.name = this.qo(a);
        c.title = this.qp(a, b);
        this.dataChart.series.add(c);
    };
    XYChart.prototype.afd = function (a) {
        this.afe(a, null);
    };
    XYChart.prototype.afe = function (a, b) {
        a(this.z2());
    };
    XYChart.prototype.afj = function (a, b) {
        a(this.z3());
    };
    XYChart.prototype.afi = function (a) {
        this.afj(a, true);
    };
    XYChart.prototype.uo = function (a, b) {
        this.afd(a);
        this.afj(a, b);
    };
    XYChart.prototype.afh = function (a) {
        this.ae8(this.z2(), a);
    };
    XYChart.prototype.afm = function (a) {
        this.ae8(this.z3(), a);
    };
    XYChart.prototype.afk = function (a) {
        this.ae7(this.z3(), a);
    };
    XYChart.prototype.aff = function (a) {
        this.ae7(this.z2(), a);
    };
    XYChart.prototype.abg = function (a) {
        var b = this.z2();
        return b.el(a, new ScalerParams(1, this.y5, this.y4, b.ch));
    };
    XYChart.prototype.abi = function (a) {
        var b = this.z2();
        return b.eo(a, new ScalerParams(1, this.y5, this.y4, b.ch));
    };
    XYChart.prototype.abh = function (a) {
        var b = this.z3();
        return b.el(a, new ScalerParams(1, this.y5, this.y4, b.ch));
    };
    XYChart.prototype.abj = function (a) {
        var b = this.z3();
        return b.eo(a, new ScalerParams(1, this.y5, this.y4, b.ch));
    };
    XYChart.prototype.wl = function (a) {
        _super.prototype.wl.call(this, a);
        if (a == null) {
            return;
        }
        if (a.axes != null) {
            var b = this.z2();
            var c = this.z3();
            if (a.axes.contains(b)) {
                a.axes.remove(b);
            }
            if (a.axes.contains(c)) {
                a.axes.remove(c);
            }
        }
    };
    XYChart.acw = function (a, b) {
        var e_1, _a;
        try {
            for (var _b = __values(fromEnum(b)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                var d = c.findMatchingHint(16);
                if (d != null) {
                    var e = d.path;
                    return e;
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        return null;
    };
    XYChart.prototype.u0 = function (a, b) {
        _super.prototype.u0.call(this, a, b);
        if (a == this.dataChart) {
            this.dataChart.axes.add(this.z2());
            this.dataChart.axes.add(this.z3());
        }
    };
    XYChart.prototype.ae9 = function (a) {
        var _this = this;
        if (a != null && this.lt != null && a.fn == null) {
            a.provideRenderer(this.lt);
        }
        a.jz = this.agc;
        a.et = this.abq;
        a.j0 = this.agd;
        a.ew = this.abr;
        a.j2 = this.agf;
        a.fi = this.abs;
        a.j1 = this.age;
        a.fj = this.abt;
        a.j3 = this.agg;
        a.fk = this.abu;
        a.c4 = true;
        a.cu = this.shouldAvoidAxisAnnotationCollisions;
        a.cw = this.shouldAvoidAxisAnnotationCollisions;
        var b = a == this.z2();
        if (b) {
            a.g8 = this.adq;
        }
        a.f2 = this.ack;
        a.gi = this.ac2;
        a.ah = this.z0;
        this.ae7(a, function (c) { return c.angle = _this.abl; });
        this.ae7(a, function (c) {
            if (c.extent != _this.abk) {
                c.extent = _this.abk;
            }
        });
        this.ae7(a, function (c) { return c.horizontalAlignment = _this.afq; });
        this.ae7(a, function (c) { return c.verticalAlignment = _this.ags; });
        this.ae7(a, function (c) { return c.visibility = _this.agx; });
        this.ae7(a, function (c) { return c.location = _this.z8(_this.aab); });
        this.ae8(a, function (c) { return c.horizontalAlignment = _this.afr; });
        this.ae8(a, function (c) { return c.angle = _this.abv; });
        this.afa(a);
    };
    XYChart.prototype.afb = function (a) {
        var _this = this;
        if (a == null) {
            return;
        }
        if (this.lt != null && a.fn == null) {
            a.provideRenderer(this.lt);
        }
        a.cu = this.shouldAvoidAxisAnnotationCollisions;
        a.cw = this.shouldAvoidAxisAnnotationCollisions;
        this.ae7(a, function (b) { return b.angle = _this.ab2; });
        this.ae7(a, function (b) {
            if (b.extent != _this.ab1) {
                b.extent = _this.ab1;
            }
        });
        this.ae7(a, function (b) { return b.horizontalAlignment = _this.afs; });
        this.ae7(a, function (b) { return b.verticalAlignment = _this.agt; });
        this.ae7(a, function (b) { return b.visibility = _this.agy; });
        this.ae7(a, function (b) { return b.location = _this.z9(_this.aad); });
        this.ae8(a, function (b) { return b.verticalAlignment = _this.agu; });
        this.ae8(a, function (b) { return b.angle = _this.acc; });
        a.jz = this.agj;
        a.et = this.ab7;
        a.j0 = this.agk;
        a.ew = this.ab8;
        a.j2 = this.agm;
        a.fi = this.ab9;
        a.j1 = this.agl;
        a.fj = this.aca;
        a.j3 = this.agn;
        a.fk = this.acb;
        var b = a == this.z3();
        if (b) {
            a.g8 = this.aev;
        }
        a.gi = this.ad7;
        a.ah = this.z1;
        this.afc(a);
    };
    XYChart.prototype.ux = function () {
        _super.prototype.ux.call(this);
        this.ae9(this.z2());
        this.afb(this.z3());
    };
    XYChart.prototype.ae7 = function (a, b) {
        if (a == null) {
            return;
        }
        if (a.az == null) {
            a.az = new AxisLabelSettings();
            a.az.angle = a.d7();
            a.az.bottomMargin = a.d8();
            a.az.leftMargin = a.d9();
            a.az.rightMargin = a.ea();
            a.az.topMargin = a.eb();
            a.az.horizontalAlignment = a.ju();
            a.az.location = a.a0();
            a.az.verticalAlignment = a.ko();
        }
        b(a.az);
    };
    XYChart.prototype.ae8 = function (a, b) {
        if (a == null) {
            return;
        }
        if (a.bf == null) {
            a.bf = new TitleSettings();
            a.bf.bottomMargin = a.ed();
            a.bf.leftMargin = a.ee();
            a.bf.rightMargin = a.ef();
            a.bf.topMargin = a.eg();
            a.bf.location = a.a1();
            a.bf.angle = a.ec();
            a.bf.horizontalAlignment = a.jv();
            a.bf.verticalAlignment = a.kp();
        }
        b(a.bf);
    };
    XYChart.prototype.vu = function (a, b, c) {
        var _this = this;
        _super.prototype.vu.call(this, a, b, c);
        switch (a) {
            case "XAxisLabelAngle":
                this.aff(function (d) { return d.angle = _this.abl; });
                break;
            case "YAxisLabelAngle":
                this.afk(function (d) { return d.angle = _this.ab2; });
                break;
            case "XAxisLabelHorizontalAlignment":
                this.aff(function (d) { return d.horizontalAlignment = _this.afq; });
                break;
            case "YAxisLabelHorizontalAlignment":
                this.afk(function (d) { return d.horizontalAlignment = _this.afs; });
                break;
            case "XAxisLabelVerticalAlignment":
                this.aff(function (d) { return d.verticalAlignment = _this.ags; });
                break;
            case "YAxisLabelVerticalAlignment":
                this.afk(function (d) { return d.verticalAlignment = _this.agt; });
                break;
            case "XAxisLabelVisibility":
                this.aff(function (d) { return d.visibility = _this.agx; });
                break;
            case "YAxisLabelVisibility":
                this.afk(function (d) { return d.visibility = _this.agy; });
                break;
            case "XAxisLabel":
                this.afd(function (d) { return d.f2 = _this.ack; });
                break;
            case "YAxisLabel":
                this.afi(function (d) { return d.f2 = _this.acl; });
                break;
            case "XAxisLabelFormat":
                this.afd(function (d) { return d.gi = _this.ac2; });
                break;
            case "XAxisLabelFormatSpecifiers":
                this.afd(function (d) { return d.ah = _this.z0; });
                break;
            case "YAxisLabelFormat":
                this.afi(function (d) { return d.gi = _this.ad7; });
                break;
            case "YAxisLabelFormatSpecifiers":
                this.afi(function (d) { return d.ah = _this.z1; });
                break;
            case "XAxisTitleAngle":
                this.afh(function (d) { return d.angle = _this.abv; });
                break;
            case "YAxisTitleAngle":
                this.afm(function (d) { return d.angle = _this.acc; });
                break;
            case "XAxisTitleAlignment":
                this.afh(function (d) { return d.horizontalAlignment = _this.afr; });
                break;
            case "YAxisTitleAlignment":
                this.afm(function (d) { return d.verticalAlignment = _this.agu; });
                break;
            case "XAxisTitle":
                this.afd(function (d) { return d.g8 = _this.adq; });
                break;
            case "YAxisTitle":
                this.afj(function (d) { return d.g8 = _this.aev; }, false);
                break;
            case "XAxisMajorStroke":
                this.afe(function (d) { return d.jz = _this.agc; }, a);
                break;
            case "YAxisMajorStroke":
                this.afi(function (d) { return d.jz = _this.agj; });
                break;
            case "XAxisMajorStrokeThickness":
                this.afe(function (d) { return d.et = _this.abq; }, a);
                break;
            case "YAxisMajorStrokeThickness":
                this.afi(function (d) { return d.et = _this.ab7; });
                break;
            case "XAxisMinorStroke":
                this.afd(function (d) { return d.j0 = _this.agd; });
                break;
            case "YAxisMinorStroke":
                this.afi(function (d) { return d.j0 = _this.agk; });
                break;
            case "XAxisMinorStrokeThickness":
                this.afd(function (d) { return d.ew = _this.abr; });
                break;
            case "YAxisMinorStrokeThickness":
                this.afi(function (d) { return d.ew = _this.ab8; });
                break;
            case "XAxisTickLength":
                this.afd(function (d) { return d.fj = _this.abt; });
                break;
            case "YAxisTickLength":
                this.afi(function (d) { return d.fj = _this.aca; });
                break;
            case "XAxisTickStroke":
                this.afd(function (d) { return d.j3 = _this.agg; });
                break;
            case "YAxisTickStroke":
                this.afi(function (d) { return d.j3 = _this.agn; });
                break;
            case "XAxisTickStrokeThickness":
                this.afd(function (d) { return d.fk = _this.abu; });
                break;
            case "YAxisTickStrokeThickness":
                this.afi(function (d) { return d.fk = _this.acb; });
                break;
            case "XAxisStroke":
                this.afd(function (d) { return d.j2 = _this.agf; });
                break;
            case "YAxisStroke":
                this.afi(function (d) { return d.j2 = _this.agm; });
                break;
            case "XAxisStrokeThickness":
                this.afd(function (d) { return d.fi = _this.abs; });
                break;
            case "YAxisStrokeThickness":
                this.afi(function (d) { return d.fi = _this.ab9; });
                break;
            case "XAxisExtent":
                this.aff(function (d) { return d.extent = _this.abk; });
                break;
            case "YAxisExtent":
                this.afk(function (d) { return d.extent = _this.ab1; });
                break;
            case "XAxisStrip":
                this.afd(function (d) { return d.j1 = _this.age; });
                break;
            case "YAxisStrip":
                this.afi(function (d) { return d.j1 = _this.agl; });
                break;
            case "XAxisInverted":
                this.afd(function (d) { return d.ch = _this.aag; });
                break;
            case "YAxisInverted":
                this.afi(function (d) { return d.ch = _this.aah; });
                break;
            case "YAxisLabelLocation":
                this.afl();
                break;
            case "XAxisLabelLocation":
                this.afg();
                break;
        }
    };
    XYChart.prototype.afl = function () {
        var a = this.z9(this.aad);
        if (a == 2 || a == 7) {
            this.afs = 2;
        }
        else {
            this.afs = 0;
        }
        this.afk(function (b) { return b.location = a; });
    };
    XYChart.prototype.afg = function () {
        var a = this.z8(this.aab);
        if (a == 1 || a == 5) {
            this.ags = 0;
        }
        else {
            this.ags = 2;
        }
        this.aff(function (b) { return b.location = a; });
    };
    XYChart.prototype.z9 = function (a) {
        if (a == 0) {
            return 2;
        }
        else if (a == 1) {
            return 3;
        }
        else if (a == 3) {
            return 7;
        }
        else if (a == 2) {
            return 6;
        }
        else {
            if (this.cb == 1) {
                return 3;
            }
            else {
                return 2;
            }
        }
    };
    XYChart.prototype.z8 = function (a) {
        if (a == 1) {
            return 1;
        }
        else if (a == 0) {
            return 0;
        }
        else if (a == 3) {
            return 5;
        }
        else if (a == 2) {
            return 4;
        }
        else {
            return 1;
        }
    };
    Object.defineProperty(XYChart.prototype, "ack", {
        get: function () {
            return this.aci;
        },
        set: function (a) {
            var b = this.ack;
            if (a != b) {
                this.aci = a;
                this.vu("XAxisLabel", b, this.ack);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acl", {
        get: function () {
            return this.acj;
        },
        set: function (a) {
            var b = this.acl;
            if (a != b) {
                this.acj = a;
                this.vu("YAxisLabel", b, this.acl);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agc", {
        get: function () {
            return this.afw;
        },
        set: function (a) {
            var b = this.agc;
            if (a != b) {
                this.afw = a;
                this.vu("XAxisMajorStroke", b, this.agc);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agj", {
        get: function () {
            return this.af3;
        },
        set: function (a) {
            var b = this.agj;
            if (a != b) {
                this.af3 = a;
                this.vu("YAxisMajorStroke", b, this.agj);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abq", {
        get: function () {
            return this.aao;
        },
        set: function (a) {
            var b = this.abq;
            if (a != b) {
                this.aao = a;
                this.vu("XAxisMajorStrokeThickness", b, this.abq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab7", {
        get: function () {
            return this.aa5;
        },
        set: function (a) {
            var b = this.ab7;
            if (a != b) {
                this.aa5 = a;
                this.vu("YAxisMajorStrokeThickness", b, this.ab7);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abr", {
        get: function () {
            return this.aap;
        },
        set: function (a) {
            var b = this.abr;
            if (a != b) {
                this.aap = a;
                this.vu("XAxisMinorStrokeThickness", b, this.abr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab8", {
        get: function () {
            return this.aa6;
        },
        set: function (a) {
            var b = this.ab8;
            if (a != b) {
                this.aa6 = a;
                this.vu("YAxisMinorStrokeThickness", b, this.ab8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "age", {
        get: function () {
            return this.afy;
        },
        set: function (a) {
            var b = this.age;
            if (a != b) {
                this.afy = a;
                this.vu("XAxisStrip", b, this.age);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agl", {
        get: function () {
            return this.af5;
        },
        set: function (a) {
            var b = this.agl;
            if (a != b) {
                this.af5 = a;
                this.vu("YAxisStrip", b, this.agl);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agf", {
        get: function () {
            return this.afz;
        },
        set: function (a) {
            var b = this.agf;
            if (a != b) {
                this.afz = a;
                this.vu("XAxisStroke", b, this.agf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agm", {
        get: function () {
            return this.af6;
        },
        set: function (a) {
            var b = this.agm;
            if (a != b) {
                this.af6 = a;
                this.vu("YAxisStroke", b, this.agm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abs", {
        get: function () {
            return this.aaq;
        },
        set: function (a) {
            var b = this.abs;
            if (a != b) {
                this.aaq = a;
                this.vu("XAxisStrokeThickness", b, this.abs);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab9", {
        get: function () {
            return this.aa7;
        },
        set: function (a) {
            var b = this.ab9;
            if (a != b) {
                this.aa7 = a;
                this.vu("YAxisStrokeThickness", b, this.ab9);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abt", {
        get: function () {
            return this.aar;
        },
        set: function (a) {
            var b = this.abt;
            if (a != b) {
                this.aar = a;
                this.vu("XAxisTickLength", b, this.abt);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aca", {
        get: function () {
            return this.aa8;
        },
        set: function (a) {
            var b = this.aca;
            if (a != b) {
                this.aa8 = a;
                this.vu("YAxisTickLength", b, this.aca);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agg", {
        get: function () {
            return this.af0;
        },
        set: function (a) {
            var b = this.agg;
            if (a != b) {
                this.af0 = a;
                this.vu("XAxisTickStroke", b, this.agg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agn", {
        get: function () {
            return this.af7;
        },
        set: function (a) {
            var b = this.agn;
            if (a != b) {
                this.af7 = a;
                this.vu("YAxisTickStroke", b, this.agn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abu", {
        get: function () {
            return this.aas;
        },
        set: function (a) {
            var b = this.abu;
            if (a != b) {
                this.aas = a;
                this.vu("XAxisTickStrokeThickness", b, this.abu);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acb", {
        get: function () {
            return this.aa9;
        },
        set: function (a) {
            var b = this.acb;
            if (a != b) {
                this.aa9 = a;
                this.vu("YAxisTickStrokeThickness", b, this.acb);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "adq", {
        get: function () {
            return this.aco;
        },
        set: function (a) {
            var b = this.adq;
            if (a != b) {
                this.aco = a;
                this.vu("XAxisTitle", b, this.adq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aev", {
        get: function () {
            return this.acs;
        },
        set: function (a) {
            var b = this.aev;
            if (a != b) {
                this.acs = a;
                this.vu("YAxisTitle", b, this.aev);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agd", {
        get: function () {
            return this.afx;
        },
        set: function (a) {
            var b = this.agd;
            if (a != b) {
                this.afx = a;
                this.vu("XAxisMinorStroke", b, this.agd);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agk", {
        get: function () {
            return this.af4;
        },
        set: function (a) {
            var b = this.agk;
            if (a != b) {
                this.af4 = a;
                this.vu("YAxisMinorStroke", b, this.agk);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abl", {
        get: function () {
            return this.aaj;
        },
        set: function (a) {
            var b = this.abl;
            if (a != b) {
                this.aaj = a;
                this.vu("XAxisLabelAngle", b, this.abl);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab2", {
        get: function () {
            return this.aa0;
        },
        set: function (a) {
            var b = this.ab2;
            if (a != b) {
                this.aa0 = a;
                this.vu("YAxisLabelAngle", b, this.ab2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abk", {
        get: function () {
            return this.aai;
        },
        set: function (a) {
            var b = this.abk;
            if (a != b) {
                this.aai = a;
                this.vu("XAxisExtent", b, this.abk);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab1", {
        get: function () {
            return this.aaz;
        },
        set: function (a) {
            var b = this.ab1;
            if (a != b) {
                this.aaz = a;
                this.vu("YAxisExtent", b, this.ab1);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abv", {
        get: function () {
            return this.aat;
        },
        set: function (a) {
            var b = this.abv;
            if (a != b) {
                this.aat = a;
                this.vu("XAxisTitleAngle", b, this.abv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "acc", {
        get: function () {
            return this.aba;
        },
        set: function (a) {
            var b = this.acc;
            if (a != b) {
                this.aba = a;
                this.vu("YAxisTitleAngle", b, this.acc);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aag", {
        get: function () {
            return this.aae;
        },
        set: function (a) {
            var b = this.aag;
            if (a != b) {
                this.aae = a;
                this.vu("XAxisInverted", b, this.aag);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aah", {
        get: function () {
            return this.aaf;
        },
        set: function (a) {
            var b = this.aah;
            if (a != b) {
                this.aaf = a;
                this.vu("YAxisInverted", b, this.aah);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "afr", {
        get: function () {
            return this.afo;
        },
        set: function (a) {
            var b = this.afr;
            if (a != b) {
                this.afo = a;
                this.vu("XAxisTitleAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.afr));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agu", {
        get: function () {
            return this.agr;
        },
        set: function (a) {
            var b = this.agu;
            if (a != b) {
                this.agr = a;
                this.vu("YAxisTitleAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.agu));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "afq", {
        get: function () {
            return this.afn;
        },
        set: function (a) {
            var b = this.afq;
            if (a != b) {
                this.afn = a;
                this.vu("XAxisLabelHorizontalAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.afq));
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.get_afs = function () {
        return this.afp;
    };
    XYChart.prototype.set_afs = function (a) {
        var b = this.afs;
        if (a != b) {
            this.afp = a;
            this.vu("YAxisLabelHorizontalAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.afs));
        }
    };
    Object.defineProperty(XYChart.prototype, "afs", {
        get: function () {
            return this.get_afs();
        },
        set: function (a) {
            this.set_afs(a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ags", {
        get: function () {
            return this.agp;
        },
        set: function (a) {
            var b = this.ags;
            if (a != b) {
                this.agp = a;
                this.vu("XAxisLabelVerticalAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.ags));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agt", {
        get: function () {
            return this.agq;
        },
        set: function (a) {
            var b = this.agt;
            if (a != b) {
                this.agq = a;
                this.vu("YAxisLabelVerticalAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.agt));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agx", {
        get: function () {
            return this.agv;
        },
        set: function (a) {
            var b = this.agx;
            if (a != b) {
                this.agv = a;
                this.vu("XAxisLabelVisibility", enumGetBox(Visibility_$type, b), enumGetBox(Visibility_$type, this.agx));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agy", {
        get: function () {
            return this.agw;
        },
        set: function (a) {
            var b = this.agy;
            if (a != b) {
                this.agw = a;
                this.vu("YAxisLabelVisibility", enumGetBox(Visibility_$type, b), enumGetBox(Visibility_$type, this.agy));
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.get_aad = function () {
        return this.aac;
    };
    XYChart.prototype.set_aad = function (a) {
        var b = this.aad;
        if (a != b) {
            this.aac = a;
            this.vu("YAxisLabelLocation", enumGetBox(YAxisLabelLocation_$type, b), enumGetBox(YAxisLabelLocation_$type, this.aad));
        }
    };
    Object.defineProperty(XYChart.prototype, "aad", {
        get: function () {
            return this.get_aad();
        },
        set: function (a) {
            this.set_aad(a);
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.get_aab = function () {
        return this.aaa;
    };
    XYChart.prototype.set_aab = function (a) {
        var b = this.aab;
        if (a != b) {
            this.aaa = a;
            this.vu("XAxisLabelLocation", enumGetBox(XAxisLabelLocation_$type, b), enumGetBox(XAxisLabelLocation_$type, this.aab));
        }
    };
    Object.defineProperty(XYChart.prototype, "aab", {
        get: function () {
            return this.get_aab();
        },
        set: function (a) {
            this.set_aab(a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ac2", {
        get: function () {
            return this.acm;
        },
        set: function (a) {
            var b = this.acm;
            if (a != b) {
                this.acm = a;
                this.vu("XAxisLabelFormat", b, this.acm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "z0", {
        get: function () {
            return this.zy;
        },
        set: function (a) {
            var b = this.zy;
            if (a != b) {
                this.zy = a;
                this.vu("XAxisLabelFormatSpecifiers", b, this.zy);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ad7", {
        get: function () {
            return this.acq;
        },
        set: function (a) {
            var b = this.acq;
            if (a != b) {
                this.acq = a;
                this.vu("YAxisLabelFormat", b, this.acq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "z1", {
        get: function () {
            return this.zz;
        },
        set: function (a) {
            var b = this.zz;
            if (a != b) {
                this.zz = a;
                this.vu("YAxisLabelFormatSpecifiers", b, this.zz);
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.$t = markType(XYChart, 'XYChart', DomainChart.$);
    return XYChart;
}(DomainChart));
export { XYChart };
