// igniteui-react-core/esm2015/lib/AggregatedDataProviderLayerInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, String_$type, markType } from "./type";
import { List$1 } from "./List$1";
import { DataSourceSummaryOperand_$type } from "./DataSourceSummaryOperand";
/**
 * @hidden
 */
export let AggregatedDataProviderLayerInfo = /*@__PURE__*/ (() => {
    class AggregatedDataProviderLayerInfo extends Base {
        constructor() {
            super(...arguments);
            this.f = new List$1(String_$type, 0);
            this.e = new List$1(String_$type, 0);
            this.h = new List$1(Base.$, 0);
            this.g = new List$1(DataSourceSummaryOperand_$type, 0);
            this.b = new List$1(String_$type, 0);
            this.c = new List$1(Base.$, 0);
            this.d = new List$1(Base.$, 0);
        }
        p(a, b, c, d) {
            this.m.add(a);
            this.l.add(b);
            this.n.add(c);
            this.o.add1(d);
        }
        get a() {
            return this.i != null && this.i.count == 1;
        }
        get m() {
            return this.f;
        }
        get l() {
            return this.e;
        }
        get o() {
            return this.h;
        }
        get n() {
            return this.g;
        }
        get i() {
            return this.b;
        }
        set i(a) {
            this.b = a;
        }
        get j() {
            return this.c;
        }
        set j(a) {
            this.c = a;
        }
        get k() {
            return this.d;
        }
        set k(a) {
            this.d = a;
        }
    }
    AggregatedDataProviderLayerInfo.$t = /*@__PURE__*/ markType(AggregatedDataProviderLayerInfo, 'AggregatedDataProviderLayerInfo');
    return AggregatedDataProviderLayerInfo;
})();
