// igniteui-react-core/esm2015/lib/AnchoredRadialSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RadialBaseDescription } from "./RadialBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let AnchoredRadialSeriesDescription = /*@__PURE__*/ (() => {
    class AnchoredRadialSeriesDescription extends RadialBaseDescription {
        get_type() {
            return "AnchoredRadialSeries";
        }
        constructor() {
            super();
            this.g7 = false;
            this.hl = null;
            this.hg = null;
            this.hj = null;
            this.hk = null;
            this.hi = null;
            this.hh = null;
            this.hf = null;
            this.g9 = 0;
            this.g5 = null;
            this.hb = 0;
            this.hc = 0;
        }
        get useCategoryNormalizedValues() {
            return this.g7;
        }
        set useCategoryNormalizedValues(a) {
            this.g7 = a;
            this.g("UseCategoryNormalizedValues");
        }
        get valueMemberPath() {
            return this.hl;
        }
        set valueMemberPath(a) {
            this.hl = a;
            this.g("ValueMemberPath");
        }
        get highlightedValueMemberPath() {
            return this.hg;
        }
        set highlightedValueMemberPath(a) {
            this.hg = a;
            this.g("HighlightedValueMemberPath");
        }
        get valueMemberAsLegendLabel() {
            return this.hj;
        }
        set valueMemberAsLegendLabel(a) {
            this.hj = a;
            this.g("ValueMemberAsLegendLabel");
        }
        get valueMemberAsLegendUnit() {
            return this.hk;
        }
        set valueMemberAsLegendUnit(a) {
            this.hk = a;
            this.g("ValueMemberAsLegendUnit");
        }
        get trendLineType() {
            return this.hi;
        }
        set trendLineType(a) {
            this.hi = a;
            this.g("TrendLineType");
        }
        get trendLineBrush() {
            return this.hh;
        }
        set trendLineBrush(a) {
            this.hh = a;
            this.g("TrendLineBrush");
        }
        get actualTrendLineBrush() {
            return this.hf;
        }
        set actualTrendLineBrush(a) {
            this.hf = a;
            this.g("ActualTrendLineBrush");
        }
        get trendLineThickness() {
            return this.g9;
        }
        set trendLineThickness(a) {
            this.g9 = a;
            this.g("TrendLineThickness");
        }
        get trendLineDashArray() {
            return this.g5;
        }
        set trendLineDashArray(a) {
            this.g5 = a;
            this.g("TrendLineDashArray");
        }
        get trendLinePeriod() {
            return this.hb;
        }
        set trendLinePeriod(a) {
            this.hb = a;
            this.g("TrendLinePeriod");
        }
        get trendLineZIndex() {
            return this.hc;
        }
        set trendLineZIndex(a) {
            this.hc = a;
            this.g("TrendLineZIndex");
        }
    }
    AnchoredRadialSeriesDescription.$t = /*@__PURE__*/ markType(AnchoredRadialSeriesDescription, 'AnchoredRadialSeriesDescription', RadialBaseDescription.$);
    return AnchoredRadialSeriesDescription;
})();
