// igniteui-react-core/esm2015/lib/AnchoredCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let AnchoredCategorySeriesDescription = /*@__PURE__*/ (() => {
    class AnchoredCategorySeriesDescription extends CategorySeriesDescription {
        get_type() {
            return "AnchoredCategorySeries";
        }
        constructor() {
            super();
            this.gl = null;
            this.gf = null;
            this.gj = null;
            this.gk = null;
            this.gi = null;
            this.gh = null;
            this.gd = null;
            this.f9 = 0;
            this.f7 = null;
            this.gb = 0;
            this.gg = null;
            this.ge = null;
        }
        get valueMemberPath() {
            return this.gl;
        }
        set valueMemberPath(a) {
            this.gl = a;
            this.g("ValueMemberPath");
        }
        get highlightedValueMemberPath() {
            return this.gf;
        }
        set highlightedValueMemberPath(a) {
            this.gf = a;
            this.g("HighlightedValueMemberPath");
        }
        get valueMemberAsLegendLabel() {
            return this.gj;
        }
        set valueMemberAsLegendLabel(a) {
            this.gj = a;
            this.g("ValueMemberAsLegendLabel");
        }
        get valueMemberAsLegendUnit() {
            return this.gk;
        }
        set valueMemberAsLegendUnit(a) {
            this.gk = a;
            this.g("ValueMemberAsLegendUnit");
        }
        get trendLineType() {
            return this.gi;
        }
        set trendLineType(a) {
            this.gi = a;
            this.g("TrendLineType");
        }
        get trendLineBrush() {
            return this.gh;
        }
        set trendLineBrush(a) {
            this.gh = a;
            this.g("TrendLineBrush");
        }
        get actualTrendLineBrush() {
            return this.gd;
        }
        set actualTrendLineBrush(a) {
            this.gd = a;
            this.g("ActualTrendLineBrush");
        }
        get trendLineThickness() {
            return this.f9;
        }
        set trendLineThickness(a) {
            this.f9 = a;
            this.g("TrendLineThickness");
        }
        get trendLineDashArray() {
            return this.f7;
        }
        set trendLineDashArray(a) {
            this.f7 = a;
            this.g("TrendLineDashArray");
        }
        get trendLinePeriod() {
            return this.gb;
        }
        set trendLinePeriod(a) {
            this.gb = a;
            this.g("TrendLinePeriod");
        }
        get markerCollisionAvoidance() {
            return this.gg;
        }
        set markerCollisionAvoidance(a) {
            this.gg = a;
            this.g("MarkerCollisionAvoidance");
        }
        get consolidatedItemHitTestBehavior() {
            return this.ge;
        }
        set consolidatedItemHitTestBehavior(a) {
            this.ge = a;
            this.g("ConsolidatedItemHitTestBehavior");
        }
    }
    AnchoredCategorySeriesDescription.$t = /*@__PURE__*/ markType(AnchoredCategorySeriesDescription, 'AnchoredCategorySeriesDescription', CategorySeriesDescription.$);
    return AnchoredCategorySeriesDescription;
})();
