// igniteui-react-core/esm2015/lib/AnimationKeyFrameEffectTiming.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let AnimationKeyFrameEffectTiming = /*@__PURE__*/ (() => {
    class AnimationKeyFrameEffectTiming extends Base {
        constructor() {
            super();
            this.f = 0;
            this.c = 0;
            this.g = 0;
            this.b = 0;
            this.h = 0;
            this.a = 0;
            this.j = 0;
            this.i = 0;
            this.onFinished = null;
            this.onCanceled = null;
            this.f = 0;
            this.c = 0;
            this.b = 3;
            this.h = 0;
            this.a = 0;
            this.j = 0;
            this.i = 1;
        }
    }
    AnimationKeyFrameEffectTiming.$t = /*@__PURE__*/ markType(AnimationKeyFrameEffectTiming, 'AnimationKeyFrameEffectTiming');
    return AnimationKeyFrameEffectTiming;
})();
