// igniteui-react-core/esm2015/lib/ArcGISOnlineMapImageryDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMapImageryDescription } from "./GeographicMapImageryDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ArcGISOnlineMapImageryDescription = /*@__PURE__*/ (() => {
    class ArcGISOnlineMapImageryDescription extends GeographicMapImageryDescription {
        get_type() {
            return "ArcGISOnlineMapImagery";
        }
        constructor() {
            super();
            this.ag = null;
            this.ak = null;
            this.ah = null;
            this.al = null;
            this.ac = false;
            this.ae = 0;
            this.aj = null;
            this.ai = null;
        }
        get mapServerUri() {
            return this.ag;
        }
        set mapServerUri(a) {
            this.ag = a;
            this.g("MapServerUri");
        }
        get userName() {
            return this.ak;
        }
        set userName(a) {
            this.ak = a;
            this.g("UserName");
        }
        get password() {
            return this.ah;
        }
        set password(a) {
            this.ah = a;
            this.g("Password");
        }
        get userToken() {
            return this.al;
        }
        set userToken(a) {
            this.al = a;
            this.g("UserToken");
        }
        get isMapPublic() {
            return this.ac;
        }
        set isMapPublic(a) {
            this.ac = a;
            this.g("IsMapPublic");
        }
        get defaultTokenTimeout() {
            return this.ae;
        }
        set defaultTokenTimeout(a) {
            this.ae = a;
            this.g("DefaultTokenTimeout");
        }
        get tokenGenerationEndPoint() {
            return this.aj;
        }
        set tokenGenerationEndPoint(a) {
            this.aj = a;
            this.g("TokenGenerationEndPoint");
        }
        get refererUri() {
            return this.ai;
        }
        set refererUri(a) {
            this.ai = a;
            this.g("RefererUri");
        }
    }
    ArcGISOnlineMapImageryDescription.$t = /*@__PURE__*/ markType(ArcGISOnlineMapImageryDescription, 'ArcGISOnlineMapImageryDescription', GeographicMapImageryDescription.$);
    return ArcGISOnlineMapImageryDescription;
})();
