// igniteui-react-core/esm2015/lib/AssigningSeriesShapeStyleEventArgsBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AssigningSeriesStyleEventArgsBaseDescription } from "./AssigningSeriesStyleEventArgsBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let AssigningSeriesShapeStyleEventArgsBaseDescription = /*@__PURE__*/ (() => {
    class AssigningSeriesShapeStyleEventArgsBaseDescription extends AssigningSeriesStyleEventArgsBaseDescription {
        get_type() {
            return "AssigningSeriesShapeStyleEventArgsBase";
        }
        constructor() {
            super();
            this.bg = 0;
            this.bc = null;
            this.be = 0;
            this.bf = 0;
        }
        get strokeThickness() {
            return this.bg;
        }
        set strokeThickness(a) {
            this.bg = a;
            this.g("StrokeThickness");
        }
        get strokeDashArray() {
            return this.bc;
        }
        set strokeDashArray(a) {
            this.bc = a;
            this.g("StrokeDashArray");
        }
        get radiusX() {
            return this.be;
        }
        set radiusX(a) {
            this.be = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.bf;
        }
        set radiusY(a) {
            this.bf = a;
            this.g("RadiusY");
        }
    }
    AssigningSeriesShapeStyleEventArgsBaseDescription.$t = /*@__PURE__*/ markType(AssigningSeriesShapeStyleEventArgsBaseDescription, 'AssigningSeriesShapeStyleEventArgsBaseDescription', AssigningSeriesStyleEventArgsBaseDescription.$);
    return AssigningSeriesShapeStyleEventArgsBaseDescription;
})();
