// igniteui-react-core/esm2015/lib/AssigningSeriesStyleEventArgsBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let AssigningSeriesStyleEventArgsBaseDescription = /*@__PURE__*/ (() => {
    class AssigningSeriesStyleEventArgsBaseDescription extends Description {
        get_type() {
            return "AssigningSeriesStyleEventArgsBase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.a2 = 0;
            this.a1 = 0;
            this.w = new Date();
            this.v = new Date();
            this.a6 = null;
            this.a5 = null;
            this.a7 = null;
            this.ad = 0;
            this.z = 0;
            this.i = null;
            this.j = null;
            this.h = null;
            this.ab = 0;
            this.af = 0;
            this.ac = 0;
            this.ag = 0;
            this.aa = 0;
            this.ae = 0;
            this.aj = 0;
            this.ak = 0;
            this.ai = 0;
            this.am = 0;
            this.al = 0;
            this.ah = 0;
            this.o = false;
            this.n = false;
            this.p = false;
            this.q = false;
        }
        get startIndex() {
            return this.a2;
        }
        set startIndex(a) {
            this.a2 = a;
            this.g("StartIndex");
        }
        get endIndex() {
            return this.a1;
        }
        set endIndex(a) {
            this.a1 = a;
            this.g("EndIndex");
        }
        get startDate() {
            return this.w;
        }
        set startDate(a) {
            this.w = a;
            this.g("StartDate");
        }
        get endDate() {
            return this.v;
        }
        set endDate(a) {
            this.v = a;
            this.g("EndDate");
        }
        get getItemsRef() {
            return this.a6;
        }
        set getItemsRef(a) {
            this.a6 = a;
            this.g("GetItemsRef");
        }
        get fill() {
            return this.a5;
        }
        set fill(a) {
            this.a5 = a;
            this.g("Fill");
        }
        get stroke() {
            return this.a7;
        }
        set stroke(a) {
            this.a7 = a;
            this.g("Stroke");
        }
        get opacity() {
            return this.ad;
        }
        set opacity(a) {
            this.ad = a;
            this.g("Opacity");
        }
        get fadeOpacity() {
            return this.z;
        }
        set fadeOpacity(a) {
            this.z = a;
            this.g("FadeOpacity");
        }
        get highlightingInfo() {
            return this.i;
        }
        set highlightingInfo(a) {
            this.i = a;
            this.g("HighlightingInfo");
        }
        get selectionHighlightingInfo() {
            return this.j;
        }
        set selectionHighlightingInfo(a) {
            this.j = a;
            this.g("SelectionHighlightingInfo");
        }
        get focusHighlightingInfo() {
            return this.h;
        }
        set focusHighlightingInfo(a) {
            this.h = a;
            this.g("FocusHighlightingInfo");
        }
        get maxAllSeriesHighlightingProgress() {
            return this.ab;
        }
        set maxAllSeriesHighlightingProgress(a) {
            this.ab = a;
            this.g("MaxAllSeriesHighlightingProgress");
        }
        get sumAllSeriesHighlightingProgress() {
            return this.af;
        }
        set sumAllSeriesHighlightingProgress(a) {
            this.af = a;
            this.g("SumAllSeriesHighlightingProgress");
        }
        get maxAllSeriesSelectionHighlightingProgress() {
            return this.ac;
        }
        set maxAllSeriesSelectionHighlightingProgress(a) {
            this.ac = a;
            this.g("MaxAllSeriesSelectionHighlightingProgress");
        }
        get sumAllSeriesSelectionHighlightingProgress() {
            return this.ag;
        }
        set sumAllSeriesSelectionHighlightingProgress(a) {
            this.ag = a;
            this.g("SumAllSeriesSelectionHighlightingProgress");
        }
        get maxAllSeriesFocusHighlightingProgress() {
            return this.aa;
        }
        set maxAllSeriesFocusHighlightingProgress(a) {
            this.aa = a;
            this.g("MaxAllSeriesFocusHighlightingProgress");
        }
        get sumAllSeriesFocusHighlightingProgress() {
            return this.ae;
        }
        set sumAllSeriesFocusHighlightingProgress(a) {
            this.ae = a;
            this.g("SumAllSeriesFocusHighlightingProgress");
        }
        get totalAllSeriesHighlightingProgress() {
            return this.aj;
        }
        set totalAllSeriesHighlightingProgress(a) {
            this.aj = a;
            this.g("TotalAllSeriesHighlightingProgress");
        }
        get totalAllSeriesHighWaterMark() {
            return this.ak;
        }
        set totalAllSeriesHighWaterMark(a) {
            this.ak = a;
            this.g("TotalAllSeriesHighWaterMark");
        }
        get totalAllSeriesFocusHighWaterMark() {
            return this.ai;
        }
        set totalAllSeriesFocusHighWaterMark(a) {
            this.ai = a;
            this.g("TotalAllSeriesFocusHighWaterMark");
        }
        get totalAllSeriesSelectionHighWaterMark() {
            return this.am;
        }
        set totalAllSeriesSelectionHighWaterMark(a) {
            this.am = a;
            this.g("TotalAllSeriesSelectionHighWaterMark");
        }
        get totalAllSeriesSelectionHighlightingProgress() {
            return this.al;
        }
        set totalAllSeriesSelectionHighlightingProgress(a) {
            this.al = a;
            this.g("TotalAllSeriesSelectionHighlightingProgress");
        }
        get totalAllSeriesFocusHighlightingProgress() {
            return this.ah;
        }
        set totalAllSeriesFocusHighlightingProgress(a) {
            this.ah = a;
            this.g("TotalAllSeriesFocusHighlightingProgress");
        }
        get highlightingHandled() {
            return this.o;
        }
        set highlightingHandled(a) {
            this.o = a;
            this.g("HighlightingHandled");
        }
        get hasDateRange() {
            return this.n;
        }
        set hasDateRange(a) {
            this.n = a;
            this.g("HasDateRange");
        }
        get isNegativeShape() {
            return this.p;
        }
        set isNegativeShape(a) {
            this.p = a;
            this.g("IsNegativeShape");
        }
        get isThumbnail() {
            return this.q;
        }
        set isThumbnail(a) {
            this.q = a;
            this.g("IsThumbnail");
        }
    }
    AssigningSeriesStyleEventArgsBaseDescription.$t = /*@__PURE__*/ markType(AssigningSeriesStyleEventArgsBaseDescription, 'AssigningSeriesStyleEventArgsBaseDescription', Description.$);
    return AssigningSeriesStyleEventArgsBaseDescription;
})();
