// igniteui-react-core/esm2015/lib/AxisDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let AxisDescription = /*@__PURE__*/ (() => {
    class AxisDescription extends Description {
        get_type() {
            return "Axis";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.by = null;
            this.cg = null;
            this.ce = null;
            this.bv = null;
            this.aw = 0;
            this.l = null;
            this.cd = null;
            this.b9 = null;
            this.bt = null;
            this.au = 0;
            this.j = null;
            this.ca = null;
            this.bu = null;
            this.av = 0;
            this.k = null;
            this.cf = null;
            this.ay = 0;
            this.m = null;
            this.ax = 0;
            this.v = false;
            this.w = false;
            this.y = false;
            this.z = false;
            this.ac = false;
            this.ab = false;
            this.bp = 0;
            this.t = false;
            this.b6 = null;
            this.b5 = null;
            this.x = false;
            this.an = 0;
            this.ap = 0;
            this.ar = 0;
            this.as = 0;
            this.at = 0;
            this.ao = 0;
            this.b4 = null;
            this.b7 = null;
            this.bx = null;
            this.br = null;
            this.cm = null;
            this.cn = null;
            this.cl = null;
            this.aa = false;
            this.az = 0;
            this.a1 = 0;
            this.a3 = 0;
            this.a4 = 0;
            this.a5 = 0;
            this.a0 = 0;
            this.ck = null;
            this.co = null;
            this.h = null;
            this.bw = null;
            this.u = false;
            this.b3 = null;
            this.i = null;
            this.cb = null;
            this.cc = null;
            this.bz = null;
            this.b0 = null;
            this.aq = 0;
            this.b2 = null;
            this.b1 = null;
            this.ch = null;
            this.a2 = 0;
            this.cj = null;
            this.ci = null;
            this.b8 = null;
        }
        get formatLabelRef() {
            return this.by;
        }
        set formatLabelRef(a) {
            this.by = a;
            this.g("FormatLabelRef");
        }
        get title() {
            return this.cg;
        }
        set title(a) {
            this.cg = a;
            this.g("Title");
        }
        get stroke() {
            return this.ce;
        }
        set stroke(a) {
            this.ce = a;
            this.g("Stroke");
        }
        get actualStroke() {
            return this.bv;
        }
        set actualStroke(a) {
            this.bv = a;
            this.g("ActualStroke");
        }
        get strokeThickness() {
            return this.aw;
        }
        set strokeThickness(a) {
            this.aw = a;
            this.g("StrokeThickness");
        }
        get strokeDashArray() {
            return this.l;
        }
        set strokeDashArray(a) {
            this.l = a;
            this.g("StrokeDashArray");
        }
        get strip() {
            return this.cd;
        }
        set strip(a) {
            this.cd = a;
            this.g("Strip");
        }
        get majorStroke() {
            return this.b9;
        }
        set majorStroke(a) {
            this.b9 = a;
            this.g("MajorStroke");
        }
        get actualMajorStroke() {
            return this.bt;
        }
        set actualMajorStroke(a) {
            this.bt = a;
            this.g("ActualMajorStroke");
        }
        get majorStrokeThickness() {
            return this.au;
        }
        set majorStrokeThickness(a) {
            this.au = a;
            this.g("MajorStrokeThickness");
        }
        get majorStrokeDashArray() {
            return this.j;
        }
        set majorStrokeDashArray(a) {
            this.j = a;
            this.g("MajorStrokeDashArray");
        }
        get minorStroke() {
            return this.ca;
        }
        set minorStroke(a) {
            this.ca = a;
            this.g("MinorStroke");
        }
        get actualMinorStroke() {
            return this.bu;
        }
        set actualMinorStroke(a) {
            this.bu = a;
            this.g("ActualMinorStroke");
        }
        get minorStrokeThickness() {
            return this.av;
        }
        set minorStrokeThickness(a) {
            this.av = a;
            this.g("MinorStrokeThickness");
        }
        get minorStrokeDashArray() {
            return this.k;
        }
        set minorStrokeDashArray(a) {
            this.k = a;
            this.g("MinorStrokeDashArray");
        }
        get tickStroke() {
            return this.cf;
        }
        set tickStroke(a) {
            this.cf = a;
            this.g("TickStroke");
        }
        get tickStrokeThickness() {
            return this.ay;
        }
        set tickStrokeThickness(a) {
            this.ay = a;
            this.g("TickStrokeThickness");
        }
        get tickStrokeDashArray() {
            return this.m;
        }
        set tickStrokeDashArray(a) {
            this.m = a;
            this.g("TickStrokeDashArray");
        }
        get tickLength() {
            return this.ax;
        }
        set tickLength(a) {
            this.ax = a;
            this.g("TickLength");
        }
        get isDisabled() {
            return this.v;
        }
        set isDisabled(a) {
            this.v = a;
            this.g("IsDisabled");
        }
        get isInverted() {
            return this.w;
        }
        set isInverted(a) {
            this.w = a;
            this.g("IsInverted");
        }
        get shouldAvoidAnnotationCollisions() {
            return this.y;
        }
        set shouldAvoidAnnotationCollisions(a) {
            this.y = a;
            this.g("ShouldAvoidAnnotationCollisions");
        }
        get shouldKeepAnnotationsInView() {
            return this.z;
        }
        set shouldKeepAnnotationsInView(a) {
            this.z = a;
            this.g("ShouldKeepAnnotationsInView");
        }
        get usePerLabelHeightMeasurement() {
            return this.ac;
        }
        set usePerLabelHeightMeasurement(a) {
            this.ac = a;
            this.g("UsePerLabelHeightMeasurement");
        }
        get useEnhancedIntervalManagement() {
            return this.ab;
        }
        set useEnhancedIntervalManagement(a) {
            this.ab = a;
            this.g("UseEnhancedIntervalManagement");
        }
        get enhancedIntervalMinimumCharacters() {
            return this.bp;
        }
        set enhancedIntervalMinimumCharacters(a) {
            this.bp = a;
            this.g("EnhancedIntervalMinimumCharacters");
        }
        get enhancedIntervalPreferMoreCategoryLabels() {
            return this.t;
        }
        set enhancedIntervalPreferMoreCategoryLabels(a) {
            this.t = a;
            this.g("EnhancedIntervalPreferMoreCategoryLabels");
        }
        get labelTextColor() {
            return this.b6;
        }
        set labelTextColor(a) {
            this.b6 = a;
            this.g("LabelTextColor");
        }
        get labelLocation() {
            return this.b5;
        }
        set labelLocation(a) {
            this.b5 = a;
            this.g("LabelLocation");
        }
        get labelShowFirstLabel() {
            return this.x;
        }
        set labelShowFirstLabel(a) {
            this.x = a;
            this.g("LabelShowFirstLabel");
        }
        get labelAngle() {
            return this.an;
        }
        set labelAngle(a) {
            this.an = a;
            this.g("LabelAngle");
        }
        get labelExtent() {
            return this.ap;
        }
        set labelExtent(a) {
            this.ap = a;
            this.g("LabelExtent");
        }
        get labelLeftMargin() {
            return this.ar;
        }
        set labelLeftMargin(a) {
            this.ar = a;
            this.g("LabelLeftMargin");
        }
        get labelRightMargin() {
            return this.as;
        }
        set labelRightMargin(a) {
            this.as = a;
            this.g("LabelRightMargin");
        }
        get labelTopMargin() {
            return this.at;
        }
        set labelTopMargin(a) {
            this.at = a;
            this.g("LabelTopMargin");
        }
        get labelBottomMargin() {
            return this.ao;
        }
        set labelBottomMargin(a) {
            this.ao = a;
            this.g("LabelBottomMargin");
        }
        get labelHorizontalAlignment() {
            return this.b4;
        }
        set labelHorizontalAlignment(a) {
            this.b4 = a;
            this.g("LabelHorizontalAlignment");
        }
        get labelVerticalAlignment() {
            return this.b7;
        }
        set labelVerticalAlignment(a) {
            this.b7 = a;
            this.g("LabelVerticalAlignment");
        }
        get crossingAxisRef() {
            return this.bx;
        }
        set crossingAxisRef(a) {
            this.bx = a;
            this.g("CrossingAxisRef");
        }
        get crossingValue() {
            return this.br;
        }
        set crossingValue(a) {
            this.br = a;
            this.g("CrossingValue");
        }
        get titlePosition() {
            return this.cm;
        }
        set titlePosition(a) {
            this.cm = a;
            this.g("TitlePosition");
        }
        get titleTextColor() {
            return this.cn;
        }
        set titleTextColor(a) {
            this.cn = a;
            this.g("TitleTextColor");
        }
        get titleLocation() {
            return this.cl;
        }
        set titleLocation(a) {
            this.cl = a;
            this.g("TitleLocation");
        }
        get titleShowFirstLabel() {
            return this.aa;
        }
        set titleShowFirstLabel(a) {
            this.aa = a;
            this.g("TitleShowFirstLabel");
        }
        get titleAngle() {
            return this.az;
        }
        set titleAngle(a) {
            this.az = a;
            this.g("TitleAngle");
        }
        get titleExtent() {
            return this.a1;
        }
        set titleExtent(a) {
            this.a1 = a;
            this.g("TitleExtent");
        }
        get titleLeftMargin() {
            return this.a3;
        }
        set titleLeftMargin(a) {
            this.a3 = a;
            this.g("TitleLeftMargin");
        }
        get titleRightMargin() {
            return this.a4;
        }
        set titleRightMargin(a) {
            this.a4 = a;
            this.g("TitleRightMargin");
        }
        get titleTopMargin() {
            return this.a5;
        }
        set titleTopMargin(a) {
            this.a5 = a;
            this.g("TitleTopMargin");
        }
        get titleBottomMargin() {
            return this.a0;
        }
        set titleBottomMargin(a) {
            this.a0 = a;
            this.g("TitleBottomMargin");
        }
        get titleHorizontalAlignment() {
            return this.ck;
        }
        set titleHorizontalAlignment(a) {
            this.ck = a;
            this.g("TitleHorizontalAlignment");
        }
        get titleVerticalAlignment() {
            return this.co;
        }
        set titleVerticalAlignment(a) {
            this.co = a;
            this.g("TitleVerticalAlignment");
        }
        get annotations() {
            return this.h;
        }
        set annotations(a) {
            this.h = a;
            this.g("Annotations");
        }
        get coercionMethodsRef() {
            return this.bw;
        }
        set coercionMethodsRef(a) {
            this.bw = a;
            this.g("CoercionMethodsRef");
        }
        get expectFunctions() {
            return this.u;
        }
        set expectFunctions(a) {
            this.u = a;
            this.g("ExpectFunctions");
        }
        get labelFormat() {
            return this.b3;
        }
        set labelFormat(a) {
            this.b3 = a;
            this.g("LabelFormat");
        }
        get labelFormatSpecifiers() {
            return this.i;
        }
        set labelFormatSpecifiers(a) {
            this.i = a;
            this.g("LabelFormatSpecifiers");
        }
        get rangeChangedRef() {
            return this.cb;
        }
        set rangeChangedRef(a) {
            this.cb = a;
            this.g("RangeChangedRef");
        }
        get renderRequestedRef() {
            return this.cc;
        }
        set renderRequestedRef(a) {
            this.cc = a;
            this.g("RenderRequestedRef");
        }
        get label() {
            return this.bz;
        }
        set label(a) {
            this.bz = a;
            this.g("Label");
        }
        get labelFontFamily() {
            return this.b0;
        }
        set labelFontFamily(a) {
            this.b0 = a;
            this.g("LabelFontFamily");
        }
        get labelFontSize() {
            return this.aq;
        }
        set labelFontSize(a) {
            this.aq = a;
            this.g("LabelFontSize");
        }
        get labelFontWeight() {
            return this.b2;
        }
        set labelFontWeight(a) {
            this.b2 = a;
            this.g("LabelFontWeight");
        }
        get labelFontStyle() {
            return this.b1;
        }
        set labelFontStyle(a) {
            this.b1 = a;
            this.g("LabelFontStyle");
        }
        get titleFontFamily() {
            return this.ch;
        }
        set titleFontFamily(a) {
            this.ch = a;
            this.g("TitleFontFamily");
        }
        get titleFontSize() {
            return this.a2;
        }
        set titleFontSize(a) {
            this.a2 = a;
            this.g("TitleFontSize");
        }
        get titleFontWeight() {
            return this.cj;
        }
        set titleFontWeight(a) {
            this.cj = a;
            this.g("TitleFontWeight");
        }
        get titleFontStyle() {
            return this.ci;
        }
        set titleFontStyle(a) {
            this.ci = a;
            this.g("TitleFontStyle");
        }
        get labelVisibility() {
            return this.b8;
        }
        set labelVisibility(a) {
            this.b8 = a;
            this.g("LabelVisibility");
        }
    }
    AxisDescription.$t = /*@__PURE__*/ markType(AxisDescription, 'AxisDescription', Description.$);
    return AxisDescription;
})();
