// igniteui-react-core/esm2015/lib/AxisAnnotationDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let AxisAnnotationDescription = /*@__PURE__*/ (() => {
    class AxisAnnotationDescription extends Description {
        get_type() {
            return "AxisAnnotation";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.ab = null;
            this.z = null;
            this.j = 0;
            this.x = null;
            this.h = null;
            this.w = null;
            this.aa = null;
            this.v = null;
            this.y = null;
            this.o = 0;
            this.l = 0;
            this.n = 0;
            this.m = 0;
            this.k = 0;
        }
        get valueRef() {
            return this.ab;
        }
        set valueRef(a) {
            this.ab = a;
            this.g("ValueRef");
        }
        get text() {
            return this.z;
        }
        set text(a) {
            this.z = a;
            this.g("Text");
        }
        get backgroundCornerRadius() {
            return this.j;
        }
        set backgroundCornerRadius(a) {
            this.j = a;
            this.g("BackgroundCornerRadius");
        }
        get labelFormat() {
            return this.x;
        }
        set labelFormat(a) {
            this.x = a;
            this.g("LabelFormat");
        }
        get labelFormatSpecifiers() {
            return this.h;
        }
        set labelFormatSpecifiers(a) {
            this.h = a;
            this.g("LabelFormatSpecifiers");
        }
        get formatLabelRef() {
            return this.w;
        }
        set formatLabelRef(a) {
            this.w = a;
            this.g("FormatLabelRef");
        }
        get textColor() {
            return this.aa;
        }
        set textColor(a) {
            this.aa = a;
            this.g("TextColor");
        }
        get background() {
            return this.v;
        }
        set background(a) {
            this.v = a;
            this.g("Background");
        }
        get outline() {
            return this.y;
        }
        set outline(a) {
            this.y = a;
            this.g("Outline");
        }
        get strokeThickness() {
            return this.o;
        }
        set strokeThickness(a) {
            this.o = a;
            this.g("StrokeThickness");
        }
        get backgroundPaddingLeft() {
            return this.l;
        }
        set backgroundPaddingLeft(a) {
            this.l = a;
            this.g("BackgroundPaddingLeft");
        }
        get backgroundPaddingTop() {
            return this.n;
        }
        set backgroundPaddingTop(a) {
            this.n = a;
            this.g("BackgroundPaddingTop");
        }
        get backgroundPaddingRight() {
            return this.m;
        }
        set backgroundPaddingRight(a) {
            this.m = a;
            this.g("BackgroundPaddingRight");
        }
        get backgroundPaddingBottom() {
            return this.k;
        }
        set backgroundPaddingBottom(a) {
            this.k = a;
            this.g("BackgroundPaddingBottom");
        }
    }
    AxisAnnotationDescription.$t = /*@__PURE__*/ markType(AxisAnnotationDescription, 'AxisAnnotationDescription', Description.$);
    return AxisAnnotationDescription;
})();
