// igniteui-react-core/esm2015/lib/AxisRangeChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let AxisRangeChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class AxisRangeChangedEventArgsDescription extends Description {
        get_type() {
            return "AxisRangeChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = 0;
            this.i = 0;
            this.j = 0;
            this.h = 0;
        }
        get oldMinimumValue() {
            return this.k;
        }
        set oldMinimumValue(a) {
            this.k = a;
            this.g("OldMinimumValue");
        }
        get minimumValue() {
            return this.i;
        }
        set minimumValue(a) {
            this.i = a;
            this.g("MinimumValue");
        }
        get oldMaximumValue() {
            return this.j;
        }
        set oldMaximumValue(a) {
            this.j = a;
            this.g("OldMaximumValue");
        }
        get maximumValue() {
            return this.h;
        }
        set maximumValue(a) {
            this.h = a;
            this.g("MaximumValue");
        }
    }
    AxisRangeChangedEventArgsDescription.$t = /*@__PURE__*/ markType(AxisRangeChangedEventArgsDescription, 'AxisRangeChangedEventArgsDescription', Description.$);
    return AxisRangeChangedEventArgsDescription;
})();
