// igniteui-react-core/esm2015/lib/BaseDataSourceDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let BaseDataSourceDescription = /*@__PURE__*/ (() => {
    class BaseDataSourceDescription extends Description {
        get_type() {
            return "BaseDataSource";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.q = false;
            this.p = false;
            this.an = 0;
            this.ao = 0;
            this.ap = 0;
            this.n = false;
            this.h = null;
            this.i = null;
            this.j = null;
            this.aw = null;
            this.r = false;
            this.t = false;
            this.o = false;
            this.v = false;
            this.u = false;
            this.s = false;
            this.x = false;
            this.w = false;
            this.y = false;
            this.z = false;
            this.av = null;
            this.au = null;
            this.at = null;
        }
        get isReadOnly() {
            return this.q;
        }
        set isReadOnly(a) {
            this.q = a;
            this.g("IsReadOnly");
        }
        get isBatchingEnabled() {
            return this.p;
        }
        set isBatchingEnabled(a) {
            this.p = a;
            this.g("IsBatchingEnabled");
        }
        get actualCount() {
            return this.an;
        }
        set actualCount(a) {
            this.an = a;
            this.g("ActualCount");
        }
        get firstVisibleIndexRequested() {
            return this.ao;
        }
        set firstVisibleIndexRequested(a) {
            this.ao = a;
            this.g("FirstVisibleIndexRequested");
        }
        get lastVisibleIndexRequested() {
            return this.ap;
        }
        set lastVisibleIndexRequested(a) {
            this.ap = a;
            this.g("LastVisibleIndexRequested");
        }
        get deferAutoRefresh() {
            return this.n;
        }
        set deferAutoRefresh(a) {
            this.n = a;
            this.g("DeferAutoRefresh");
        }
        get primaryKey() {
            return this.h;
        }
        set primaryKey(a) {
            this.h = a;
            this.g("PrimaryKey");
        }
        get propertiesRequested() {
            return this.i;
        }
        set propertiesRequested(a) {
            this.i = a;
            this.g("PropertiesRequested");
        }
        get schemaIncludedProperties() {
            return this.j;
        }
        set schemaIncludedProperties(a) {
            this.j = a;
            this.g("SchemaIncludedProperties");
        }
        get sectionHeaderDisplayMode() {
            return this.aw;
        }
        set sectionHeaderDisplayMode(a) {
            this.aw = a;
            this.g("SectionHeaderDisplayMode");
        }
        get isSectionCollapsable() {
            return this.r;
        }
        set isSectionCollapsable(a) {
            this.r = a;
            this.g("IsSectionCollapsable");
        }
        get isSectionExpandedDefault() {
            return this.t;
        }
        set isSectionExpandedDefault(a) {
            this.t = a;
            this.g("IsSectionExpandedDefault");
        }
        get includeSummaryRowsInSection() {
            return this.o;
        }
        set includeSummaryRowsInSection(a) {
            this.o = a;
            this.g("IncludeSummaryRowsInSection");
        }
        get isSectionSummaryRowsAtBottom() {
            return this.v;
        }
        set isSectionSummaryRowsAtBottom(a) {
            this.v = a;
            this.g("IsSectionSummaryRowsAtBottom");
        }
        get isSectionHeaderNormalRow() {
            return this.u;
        }
        set isSectionHeaderNormalRow(a) {
            this.u = a;
            this.g("IsSectionHeaderNormalRow");
        }
        get isSectionContentVisible() {
            return this.s;
        }
        set isSectionContentVisible(a) {
            this.s = a;
            this.g("IsSectionContentVisible");
        }
        get shouldEmitSectionHeaders() {
            return this.x;
        }
        set shouldEmitSectionHeaders(a) {
            this.x = a;
            this.g("ShouldEmitSectionHeaders");
        }
        get shouldEmitSectionFooters() {
            return this.w;
        }
        set shouldEmitSectionFooters(a) {
            this.w = a;
            this.g("ShouldEmitSectionFooters");
        }
        get shouldEmitShiftedRows() {
            return this.y;
        }
        set shouldEmitShiftedRows(a) {
            this.y = a;
            this.g("ShouldEmitShiftedRows");
        }
        get shouldEmitSummaryRows() {
            return this.z;
        }
        set shouldEmitSummaryRows(a) {
            this.z = a;
            this.g("ShouldEmitSummaryRows");
        }
        get schemaChangedRef() {
            return this.av;
        }
        set schemaChangedRef(a) {
            this.av = a;
            this.g("SchemaChangedRef");
        }
        get rowExpansionChangedRef() {
            return this.au;
        }
        set rowExpansionChangedRef(a) {
            this.au = a;
            this.g("RowExpansionChangedRef");
        }
        get rootSummariesChangedRef() {
            return this.at;
        }
        set rootSummariesChangedRef(a) {
            this.at = a;
            this.g("RootSummariesChangedRef");
        }
    }
    BaseDataSourceDescription.$t = /*@__PURE__*/ markType(BaseDataSourceDescription, 'BaseDataSourceDescription', Description.$);
    return BaseDataSourceDescription;
})();
