// igniteui-react-core/esm2015/lib/BingMapsMapImageryDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMapImageryDescription } from "./GeographicMapImageryDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let BingMapsMapImageryDescription = /*@__PURE__*/ (() => {
    class BingMapsMapImageryDescription extends GeographicMapImageryDescription {
        get_type() {
            return "BingMapsMapImagery";
        }
        constructor() {
            super();
            this.ah = false;
            this.ag = false;
            this.aq = null;
            this.al = null;
            this.ad = null;
            this.ac = null;
            this.an = null;
            this.ak = null;
            this.ao = null;
            this.am = null;
            this.ap = null;
        }
        get isInitialized() {
            return this.ah;
        }
        set isInitialized(a) {
            this.ah = a;
            this.g("IsInitialized");
        }
        get isDeferredLoad() {
            return this.ag;
        }
        set isDeferredLoad(a) {
            this.ag = a;
            this.g("IsDeferredLoad");
        }
        get tilePath() {
            return this.aq;
        }
        set tilePath(a) {
            this.aq = a;
            this.g("TilePath");
        }
        get actualTilePath() {
            return this.al;
        }
        set actualTilePath(a) {
            this.al = a;
            this.g("ActualTilePath");
        }
        get subDomains() {
            return this.ad;
        }
        set subDomains(a) {
            this.ad = a;
            this.g("SubDomains");
        }
        get actualSubDomains() {
            return this.ac;
        }
        set actualSubDomains(a) {
            this.ac = a;
            this.g("ActualSubDomains");
        }
        get bingImageryRestUri() {
            return this.an;
        }
        set bingImageryRestUri(a) {
            this.an = a;
            this.g("BingImageryRestUri");
        }
        get actualBingImageryRestUri() {
            return this.ak;
        }
        set actualBingImageryRestUri(a) {
            this.ak = a;
            this.g("ActualBingImageryRestUri");
        }
        get cultureName() {
            return this.ao;
        }
        set cultureName(a) {
            this.ao = a;
            this.g("CultureName");
        }
        get apiKey() {
            return this.am;
        }
        set apiKey(a) {
            this.am = a;
            this.g("ApiKey");
        }
        get imageryStyle() {
            return this.ap;
        }
        set imageryStyle(a) {
            this.ap = a;
            this.g("ImageryStyle");
        }
    }
    BingMapsMapImageryDescription.$t = /*@__PURE__*/ markType(BingMapsMapImageryDescription, 'BingMapsMapImageryDescription', GeographicMapImageryDescription.$);
    return BingMapsMapImageryDescription;
})();
