// igniteui-react-core/esm2015/lib/BlazorCodeGeneratingComponentRendererModuleEmitter.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DotNetCodeGeneratingComponentRendererModuleEmitter } from "./DotNetCodeGeneratingComponentRendererModuleEmitter";
import { markType } from "./type";
/**
 * @hidden
 */
export let BlazorCodeGeneratingComponentRendererModuleEmitter = /*@__PURE__*/ (() => {
    class BlazorCodeGeneratingComponentRendererModuleEmitter extends DotNetCodeGeneratingComponentRendererModuleEmitter {
        constructor(a, b, c) {
            super(a, b, c);
            this.w = null;
            this.x = 0;
        }
        p() {
            super.p();
        }
        m(a, b, c) {
            super.m(a, b, c);
            if (this.x > 0) {
                this.d.l(", ");
            }
            let d = "";
            let e = b.split('/');
            if (e.length == 1) {
                d = e[0];
            }
            else {
                d = e[1];
            }
            if (d == "withDescriptions") {
                this.d.j("typeof(IgbInputModule)");
            }
            else {
                this.d.j("typeof(");
                this.d.j("Igb" + d);
                this.d.j(")");
            }
            this.x++;
        }
        o() {
            super.o();
        }
    }
    BlazorCodeGeneratingComponentRendererModuleEmitter.$t = /*@__PURE__*/ markType(BlazorCodeGeneratingComponentRendererModuleEmitter, 'BlazorCodeGeneratingComponentRendererModuleEmitter', DotNetCodeGeneratingComponentRendererModuleEmitter.$);
    return BlazorCodeGeneratingComponentRendererModuleEmitter;
})();
