// igniteui-react-core/esm2015/lib/BoxCollisionGeometry.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CollisionGeometry } from "./CollisionGeometry";
import { markType } from "./type";
/**
 * @hidden
 */
export let BoxCollisionGeometry = /*@__PURE__*/ (() => {
    class BoxCollisionGeometry extends CollisionGeometry {
        get_type() {
            return 0;
        }
        get type() {
            return this.get_type();
        }
        constructor(a) {
            super();
            this.g = null;
            this.g = a;
        }
        get_boundingBox() {
            return this.g;
        }
        get boundingBox() {
            return this.get_boundingBox();
        }
        d(a) {
            return true;
        }
    }
    BoxCollisionGeometry.$t = /*@__PURE__*/ markType(BoxCollisionGeometry, 'BoxCollisionGeometry', CollisionGeometry.$);
    return BoxCollisionGeometry;
})();
