// igniteui-react-core/esm2015/lib/ButtonDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ButtonDescription = /*@__PURE__*/ (() => {
    class ButtonDescription extends Description {
        get_type() {
            return "Button";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.a8 = 0;
            this.a7 = 0;
            this.e3 = null;
            this.ey = null;
            this.ev = null;
            this.ef = null;
            this.ee = null;
            this.er = null;
            this.eo = null;
            this.ec = null;
            this.cs = 0;
            this.cq = 0;
            this.cr = 0;
            this.ep = null;
            this.af = 0;
            this.ae = 0;
            this.ag = 0;
            this.ah = 0;
            this.ed = null;
            this.eg = null;
            this.eh = null;
            this.ei = null;
            this.cp = 0;
            this.ej = null;
            this.eq = null;
            this.en = null;
            this.ek = null;
            this.el = null;
            this.em = null;
            this.ad = 0;
            this.gs = null;
            this.gu = null;
            this.gv = null;
            this.gw = null;
            this.dc = 0;
            this.gz = null;
            this.gt = null;
            this.df = 0;
            this.de = 0;
            this.dd = 0;
            this.be = 0;
            this.bg = 0;
            this.bf = 0;
            this.bh = 0;
            this.bi = 0;
            this.g1 = null;
            this.fg = null;
            this.fi = null;
            this.fj = null;
            this.fk = null;
            this.cy = 0;
            this.fn = null;
            this.fh = null;
            this.c1 = 0;
            this.c0 = 0;
            this.cz = 0;
            this.fp = null;
            this.fq = null;
            this.fo = null;
            this.fl = null;
            this.fm = null;
            this.aw = 0;
            this.ay = 0;
            this.ax = 0;
            this.az = 0;
            this.a0 = 0;
            this.eu = null;
            this.ez = null;
            this.e0 = null;
            this.e1 = null;
            this.ct = 0;
            this.e4 = null;
            this.f0 = null;
            this.ew = null;
            this.dg = 0;
            this.c3 = 0;
            this.c2 = 0;
            this.fv = null;
            this.a1 = 0;
            this.fw = null;
            this.fx = null;
            this.g4 = null;
            this.f1 = null;
            this.g3 = null;
            this.g2 = null;
            this.g0 = null;
            this.gx = null;
            this.gy = null;
            this.ft = null;
            this.fu = null;
            this.ai = 0;
            this.ao = 0;
            this.an = 0;
            this.ap = 0;
            this.aq = 0;
            this.gh = null;
            this.gj = null;
            this.gk = null;
            this.gl = null;
            this.c8 = 0;
            this.go = null;
            this.gi = null;
            this.db = 0;
            this.da = 0;
            this.c9 = 0;
            this.a9 = 0;
            this.bb = 0;
            this.ba = 0;
            this.bc = 0;
            this.bd = 0;
            this.gq = null;
            this.gr = null;
            this.gp = null;
            this.gm = null;
            this.gn = null;
            this.e5 = null;
            this.e7 = null;
            this.e8 = null;
            this.e9 = null;
            this.cu = 0;
            this.fc = null;
            this.e6 = null;
            this.cx = 0;
            this.cw = 0;
            this.cv = 0;
            this.ar = 0;
            this.at = 0;
            this.as = 0;
            this.au = 0;
            this.av = 0;
            this.fe = null;
            this.ff = null;
            this.fd = null;
            this.fa = null;
            this.fb = null;
            this.f2 = null;
            this.f4 = null;
            this.f5 = null;
            this.f6 = null;
            this.c4 = 0;
            this.f9 = null;
            this.f3 = null;
            this.c7 = 0;
            this.c6 = 0;
            this.c5 = 0;
            this.a2 = 0;
            this.a4 = 0;
            this.a3 = 0;
            this.a5 = 0;
            this.a6 = 0;
            this.gb = null;
            this.gc = null;
            this.ga = null;
            this.f7 = null;
            this.f8 = null;
            this.ge = null;
            this.gd = null;
            this.e2 = null;
            this.fr = null;
            this.fs = null;
            this.es = null;
            this.g5 = null;
            this.gg = null;
            this.dh = 0;
            this.et = null;
            this.o = false;
            this.p = false;
            this.q = false;
            this.m = false;
            this.h = false;
            this.k = false;
            this.r = false;
            this.j = false;
            this.n = false;
            this.l = false;
            this.ak = 0;
            this.am = 0;
            this.al = 0;
            this.aj = 0;
            this.fz = null;
            this.g6 = null;
            this.i = false;
            this.ex = null;
            this.fy = null;
            this.gf = null;
        }
        get minWidth() {
            return this.a8;
        }
        set minWidth(a) {
            this.a8 = a;
            this.g("MinWidth");
        }
        get minHeight() {
            return this.a7;
        }
        set minHeight(a) {
            this.a7 = a;
            this.g("MinHeight");
        }
        get displayType() {
            return this.e3;
        }
        set displayType(a) {
            this.e3 = a;
            this.g("DisplayType");
        }
        get density() {
            return this.ey;
        }
        set density(a) {
            this.ey = a;
            this.g("Density");
        }
        get baseTheme() {
            return this.ev;
        }
        set baseTheme(a) {
            this.ev = a;
            this.g("BaseTheme");
        }
        get actualDensity() {
            return this.ef;
        }
        set actualDensity(a) {
            this.ef = a;
            this.g("ActualDensity");
        }
        get actualBorderColor() {
            return this.ee;
        }
        set actualBorderColor(a) {
            this.ee = a;
            this.g("ActualBorderColor");
        }
        get actualUmbraShadowColor() {
            return this.er;
        }
        set actualUmbraShadowColor(a) {
            this.er = a;
            this.g("ActualUmbraShadowColor");
        }
        get actualPenumbraShadowColor() {
            return this.eo;
        }
        set actualPenumbraShadowColor(a) {
            this.eo = a;
            this.g("ActualPenumbraShadowColor");
        }
        get actualAmbientShadowColor() {
            return this.ec;
        }
        set actualAmbientShadowColor(a) {
            this.ec = a;
            this.g("ActualAmbientShadowColor");
        }
        get actualRestingElevation() {
            return this.cs;
        }
        set actualRestingElevation(a) {
            this.cs = a;
            this.g("ActualRestingElevation");
        }
        get actualFocusElevation() {
            return this.cq;
        }
        set actualFocusElevation(a) {
            this.cq = a;
            this.g("ActualFocusElevation");
        }
        get actualHoverElevation() {
            return this.cr;
        }
        set actualHoverElevation(a) {
            this.cr = a;
            this.g("ActualHoverElevation");
        }
        get actualRippleColor() {
            return this.ep;
        }
        set actualRippleColor(a) {
            this.ep = a;
            this.g("ActualRippleColor");
        }
        get actualCornerRadiusBottomRight() {
            return this.af;
        }
        set actualCornerRadiusBottomRight(a) {
            this.af = a;
            this.g("ActualCornerRadiusBottomRight");
        }
        get actualCornerRadiusBottomLeft() {
            return this.ae;
        }
        set actualCornerRadiusBottomLeft(a) {
            this.ae = a;
            this.g("ActualCornerRadiusBottomLeft");
        }
        get actualCornerRadiusTopLeft() {
            return this.ag;
        }
        set actualCornerRadiusTopLeft(a) {
            this.ag = a;
            this.g("ActualCornerRadiusTopLeft");
        }
        get actualCornerRadiusTopRight() {
            return this.ah;
        }
        set actualCornerRadiusTopRight(a) {
            this.ah = a;
            this.g("ActualCornerRadiusTopRight");
        }
        get actualBackgroundColor() {
            return this.ed;
        }
        set actualBackgroundColor(a) {
            this.ed = a;
            this.g("ActualBackgroundColor");
        }
        get actualDisabledBackgroundColor() {
            return this.eg;
        }
        set actualDisabledBackgroundColor(a) {
            this.eg = a;
            this.g("ActualDisabledBackgroundColor");
        }
        get actualDisabledBorderColor() {
            return this.eh;
        }
        set actualDisabledBorderColor(a) {
            this.eh = a;
            this.g("ActualDisabledBorderColor");
        }
        get actualDisabledTextColor() {
            return this.ei;
        }
        set actualDisabledTextColor(a) {
            this.ei = a;
            this.g("ActualDisabledTextColor");
        }
        get actualDisabledElevation() {
            return this.cp;
        }
        set actualDisabledElevation(a) {
            this.cp = a;
            this.g("ActualDisabledElevation");
        }
        get actualElevationMode() {
            return this.ej;
        }
        set actualElevationMode(a) {
            this.ej = a;
            this.g("ActualElevationMode");
        }
        get actualTextColor() {
            return this.eq;
        }
        set actualTextColor(a) {
            this.eq = a;
            this.g("ActualTextColor");
        }
        get actualHoverTextColor() {
            return this.en;
        }
        set actualHoverTextColor(a) {
            this.en = a;
            this.g("ActualHoverTextColor");
        }
        get actualFocusBackgroundColor() {
            return this.ek;
        }
        set actualFocusBackgroundColor(a) {
            this.ek = a;
            this.g("ActualFocusBackgroundColor");
        }
        get actualFocusTextColor() {
            return this.el;
        }
        set actualFocusTextColor(a) {
            this.el = a;
            this.g("ActualFocusTextColor");
        }
        get actualHoverBackgroundColor() {
            return this.em;
        }
        set actualHoverBackgroundColor(a) {
            this.em = a;
            this.g("ActualHoverBackgroundColor");
        }
        get actualBorderWidth() {
            return this.ad;
        }
        set actualBorderWidth(a) {
            this.ad = a;
            this.g("ActualBorderWidth");
        }
        get raisedBackgroundColor() {
            return this.gs;
        }
        set raisedBackgroundColor(a) {
            this.gs = a;
            this.g("RaisedBackgroundColor");
        }
        get raisedDisabledBackgroundColor() {
            return this.gu;
        }
        set raisedDisabledBackgroundColor(a) {
            this.gu = a;
            this.g("RaisedDisabledBackgroundColor");
        }
        get raisedDisabledBorderColor() {
            return this.gv;
        }
        set raisedDisabledBorderColor(a) {
            this.gv = a;
            this.g("RaisedDisabledBorderColor");
        }
        get raisedDisabledTextColor() {
            return this.gw;
        }
        set raisedDisabledTextColor(a) {
            this.gw = a;
            this.g("RaisedDisabledTextColor");
        }
        get raisedDisabledElevation() {
            return this.dc;
        }
        set raisedDisabledElevation(a) {
            this.dc = a;
            this.g("RaisedDisabledElevation");
        }
        get raisedHoverBackgroundColor() {
            return this.gz;
        }
        set raisedHoverBackgroundColor(a) {
            this.gz = a;
            this.g("RaisedHoverBackgroundColor");
        }
        get raisedBorderColor() {
            return this.gt;
        }
        set raisedBorderColor(a) {
            this.gt = a;
            this.g("RaisedBorderColor");
        }
        get raisedRestingElevation() {
            return this.df;
        }
        set raisedRestingElevation(a) {
            this.df = a;
            this.g("RaisedRestingElevation");
        }
        get raisedHoverElevation() {
            return this.de;
        }
        set raisedHoverElevation(a) {
            this.de = a;
            this.g("RaisedHoverElevation");
        }
        get raisedFocusElevation() {
            return this.dd;
        }
        set raisedFocusElevation(a) {
            this.dd = a;
            this.g("RaisedFocusElevation");
        }
        get raisedBorderWidth() {
            return this.be;
        }
        set raisedBorderWidth(a) {
            this.be = a;
            this.g("RaisedBorderWidth");
        }
        get raisedCornerRadiusBottomRight() {
            return this.bg;
        }
        set raisedCornerRadiusBottomRight(a) {
            this.bg = a;
            this.g("RaisedCornerRadiusBottomRight");
        }
        get raisedCornerRadiusBottomLeft() {
            return this.bf;
        }
        set raisedCornerRadiusBottomLeft(a) {
            this.bf = a;
            this.g("RaisedCornerRadiusBottomLeft");
        }
        get raisedCornerRadiusTopLeft() {
            return this.bh;
        }
        set raisedCornerRadiusTopLeft(a) {
            this.bh = a;
            this.g("RaisedCornerRadiusTopLeft");
        }
        get raisedCornerRadiusTopRight() {
            return this.bi;
        }
        set raisedCornerRadiusTopRight(a) {
            this.bi = a;
            this.g("RaisedCornerRadiusTopRight");
        }
        get raisedRippleColor() {
            return this.g1;
        }
        set raisedRippleColor(a) {
            this.g1 = a;
            this.g("RaisedRippleColor");
        }
        get flatBackgroundColor() {
            return this.fg;
        }
        set flatBackgroundColor(a) {
            this.fg = a;
            this.g("FlatBackgroundColor");
        }
        get flatDisabledBackgroundColor() {
            return this.fi;
        }
        set flatDisabledBackgroundColor(a) {
            this.fi = a;
            this.g("FlatDisabledBackgroundColor");
        }
        get flatDisabledBorderColor() {
            return this.fj;
        }
        set flatDisabledBorderColor(a) {
            this.fj = a;
            this.g("FlatDisabledBorderColor");
        }
        get flatDisabledTextColor() {
            return this.fk;
        }
        set flatDisabledTextColor(a) {
            this.fk = a;
            this.g("FlatDisabledTextColor");
        }
        get flatDisabledElevation() {
            return this.cy;
        }
        set flatDisabledElevation(a) {
            this.cy = a;
            this.g("FlatDisabledElevation");
        }
        get flatHoverBackgroundColor() {
            return this.fn;
        }
        set flatHoverBackgroundColor(a) {
            this.fn = a;
            this.g("FlatHoverBackgroundColor");
        }
        get flatBorderColor() {
            return this.fh;
        }
        set flatBorderColor(a) {
            this.fh = a;
            this.g("FlatBorderColor");
        }
        get flatRestingElevation() {
            return this.c1;
        }
        set flatRestingElevation(a) {
            this.c1 = a;
            this.g("FlatRestingElevation");
        }
        get flatHoverElevation() {
            return this.c0;
        }
        set flatHoverElevation(a) {
            this.c0 = a;
            this.g("FlatHoverElevation");
        }
        get flatFocusElevation() {
            return this.cz;
        }
        set flatFocusElevation(a) {
            this.cz = a;
            this.g("FlatFocusElevation");
        }
        get flatRippleColor() {
            return this.fp;
        }
        set flatRippleColor(a) {
            this.fp = a;
            this.g("FlatRippleColor");
        }
        get flatTextColor() {
            return this.fq;
        }
        set flatTextColor(a) {
            this.fq = a;
            this.g("FlatTextColor");
        }
        get flatHoverTextColor() {
            return this.fo;
        }
        set flatHoverTextColor(a) {
            this.fo = a;
            this.g("FlatHoverTextColor");
        }
        get flatFocusBackgroundColor() {
            return this.fl;
        }
        set flatFocusBackgroundColor(a) {
            this.fl = a;
            this.g("FlatFocusBackgroundColor");
        }
        get flatFocusTextColor() {
            return this.fm;
        }
        set flatFocusTextColor(a) {
            this.fm = a;
            this.g("FlatFocusTextColor");
        }
        get flatBorderWidth() {
            return this.aw;
        }
        set flatBorderWidth(a) {
            this.aw = a;
            this.g("FlatBorderWidth");
        }
        get flatCornerRadiusBottomRight() {
            return this.ay;
        }
        set flatCornerRadiusBottomRight(a) {
            this.ay = a;
            this.g("FlatCornerRadiusBottomRight");
        }
        get flatCornerRadiusBottomLeft() {
            return this.ax;
        }
        set flatCornerRadiusBottomLeft(a) {
            this.ax = a;
            this.g("FlatCornerRadiusBottomLeft");
        }
        get flatCornerRadiusTopLeft() {
            return this.az;
        }
        set flatCornerRadiusTopLeft(a) {
            this.az = a;
            this.g("FlatCornerRadiusTopLeft");
        }
        get flatCornerRadiusTopRight() {
            return this.a0;
        }
        set flatCornerRadiusTopRight(a) {
            this.a0 = a;
            this.g("FlatCornerRadiusTopRight");
        }
        get backgroundColor() {
            return this.eu;
        }
        set backgroundColor(a) {
            this.eu = a;
            this.g("BackgroundColor");
        }
        get disabledBackgroundColor() {
            return this.ez;
        }
        set disabledBackgroundColor(a) {
            this.ez = a;
            this.g("DisabledBackgroundColor");
        }
        get disabledBorderColor() {
            return this.e0;
        }
        set disabledBorderColor(a) {
            this.e0 = a;
            this.g("DisabledBorderColor");
        }
        get disabledTextColor() {
            return this.e1;
        }
        set disabledTextColor(a) {
            this.e1 = a;
            this.g("DisabledTextColor");
        }
        get disabledElevation() {
            return this.ct;
        }
        set disabledElevation(a) {
            this.ct = a;
            this.g("DisabledElevation");
        }
        get elevationMode() {
            return this.e4;
        }
        set elevationMode(a) {
            this.e4 = a;
            this.g("ElevationMode");
        }
        get hoverBackgroundColor() {
            return this.f0;
        }
        set hoverBackgroundColor(a) {
            this.f0 = a;
            this.g("HoverBackgroundColor");
        }
        get borderColor() {
            return this.ew;
        }
        set borderColor(a) {
            this.ew = a;
            this.g("BorderColor");
        }
        get restingElevation() {
            return this.dg;
        }
        set restingElevation(a) {
            this.dg = a;
            this.g("RestingElevation");
        }
        get hoverElevation() {
            return this.c3;
        }
        set hoverElevation(a) {
            this.c3 = a;
            this.g("HoverElevation");
        }
        get focusElevation() {
            return this.c2;
        }
        set focusElevation(a) {
            this.c2 = a;
            this.g("FocusElevation");
        }
        get fontFamily() {
            return this.fv;
        }
        set fontFamily(a) {
            this.fv = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.a1;
        }
        set fontSize(a) {
            this.a1 = a;
            this.g("FontSize");
        }
        get fontStyle() {
            return this.fw;
        }
        set fontStyle(a) {
            this.fw = a;
            this.g("FontStyle");
        }
        get fontWeight() {
            return this.fx;
        }
        set fontWeight(a) {
            this.fx = a;
            this.g("FontWeight");
        }
        get textColor() {
            return this.g4;
        }
        set textColor(a) {
            this.g4 = a;
            this.g("TextColor");
        }
        get hoverTextColor() {
            return this.f1;
        }
        set hoverTextColor(a) {
            this.f1 = a;
            this.g("HoverTextColor");
        }
        get rippleColor() {
            return this.g3;
        }
        set rippleColor(a) {
            this.g3 = a;
            this.g("RippleColor");
        }
        get raisedTextColor() {
            return this.g2;
        }
        set raisedTextColor(a) {
            this.g2 = a;
            this.g("RaisedTextColor");
        }
        get raisedHoverTextColor() {
            return this.g0;
        }
        set raisedHoverTextColor(a) {
            this.g0 = a;
            this.g("RaisedHoverTextColor");
        }
        get raisedFocusBackgroundColor() {
            return this.gx;
        }
        set raisedFocusBackgroundColor(a) {
            this.gx = a;
            this.g("RaisedFocusBackgroundColor");
        }
        get raisedFocusTextColor() {
            return this.gy;
        }
        set raisedFocusTextColor(a) {
            this.gy = a;
            this.g("RaisedFocusTextColor");
        }
        get focusBackgroundColor() {
            return this.ft;
        }
        set focusBackgroundColor(a) {
            this.ft = a;
            this.g("FocusBackgroundColor");
        }
        get focusTextColor() {
            return this.fu;
        }
        set focusTextColor(a) {
            this.fu = a;
            this.g("FocusTextColor");
        }
        get borderWidth() {
            return this.ai;
        }
        set borderWidth(a) {
            this.ai = a;
            this.g("BorderWidth");
        }
        get cornerRadiusBottomRight() {
            return this.ao;
        }
        set cornerRadiusBottomRight(a) {
            this.ao = a;
            this.g("CornerRadiusBottomRight");
        }
        get cornerRadiusBottomLeft() {
            return this.an;
        }
        set cornerRadiusBottomLeft(a) {
            this.an = a;
            this.g("CornerRadiusBottomLeft");
        }
        get cornerRadiusTopLeft() {
            return this.ap;
        }
        set cornerRadiusTopLeft(a) {
            this.ap = a;
            this.g("CornerRadiusTopLeft");
        }
        get cornerRadiusTopRight() {
            return this.aq;
        }
        set cornerRadiusTopRight(a) {
            this.aq = a;
            this.g("CornerRadiusTopRight");
        }
        get outlinedBackgroundColor() {
            return this.gh;
        }
        set outlinedBackgroundColor(a) {
            this.gh = a;
            this.g("OutlinedBackgroundColor");
        }
        get outlinedDisabledBackgroundColor() {
            return this.gj;
        }
        set outlinedDisabledBackgroundColor(a) {
            this.gj = a;
            this.g("OutlinedDisabledBackgroundColor");
        }
        get outlinedDisabledBorderColor() {
            return this.gk;
        }
        set outlinedDisabledBorderColor(a) {
            this.gk = a;
            this.g("OutlinedDisabledBorderColor");
        }
        get outlinedDisabledTextColor() {
            return this.gl;
        }
        set outlinedDisabledTextColor(a) {
            this.gl = a;
            this.g("OutlinedDisabledTextColor");
        }
        get outlinedDisabledElevation() {
            return this.c8;
        }
        set outlinedDisabledElevation(a) {
            this.c8 = a;
            this.g("OutlinedDisabledElevation");
        }
        get outlinedHoverBackgroundColor() {
            return this.go;
        }
        set outlinedHoverBackgroundColor(a) {
            this.go = a;
            this.g("OutlinedHoverBackgroundColor");
        }
        get outlinedBorderColor() {
            return this.gi;
        }
        set outlinedBorderColor(a) {
            this.gi = a;
            this.g("OutlinedBorderColor");
        }
        get outlinedRestingElevation() {
            return this.db;
        }
        set outlinedRestingElevation(a) {
            this.db = a;
            this.g("OutlinedRestingElevation");
        }
        get outlinedHoverElevation() {
            return this.da;
        }
        set outlinedHoverElevation(a) {
            this.da = a;
            this.g("OutlinedHoverElevation");
        }
        get outlinedFocusElevation() {
            return this.c9;
        }
        set outlinedFocusElevation(a) {
            this.c9 = a;
            this.g("OutlinedFocusElevation");
        }
        get outlinedBorderWidth() {
            return this.a9;
        }
        set outlinedBorderWidth(a) {
            this.a9 = a;
            this.g("OutlinedBorderWidth");
        }
        get outlinedCornerRadiusBottomRight() {
            return this.bb;
        }
        set outlinedCornerRadiusBottomRight(a) {
            this.bb = a;
            this.g("OutlinedCornerRadiusBottomRight");
        }
        get outlinedCornerRadiusBottomLeft() {
            return this.ba;
        }
        set outlinedCornerRadiusBottomLeft(a) {
            this.ba = a;
            this.g("OutlinedCornerRadiusBottomLeft");
        }
        get outlinedCornerRadiusTopLeft() {
            return this.bc;
        }
        set outlinedCornerRadiusTopLeft(a) {
            this.bc = a;
            this.g("OutlinedCornerRadiusTopLeft");
        }
        get outlinedCornerRadiusTopRight() {
            return this.bd;
        }
        set outlinedCornerRadiusTopRight(a) {
            this.bd = a;
            this.g("OutlinedCornerRadiusTopRight");
        }
        get outlinedRippleColor() {
            return this.gq;
        }
        set outlinedRippleColor(a) {
            this.gq = a;
            this.g("OutlinedRippleColor");
        }
        get outlinedTextColor() {
            return this.gr;
        }
        set outlinedTextColor(a) {
            this.gr = a;
            this.g("OutlinedTextColor");
        }
        get outlinedHoverTextColor() {
            return this.gp;
        }
        set outlinedHoverTextColor(a) {
            this.gp = a;
            this.g("OutlinedHoverTextColor");
        }
        get outlinedFocusBackgroundColor() {
            return this.gm;
        }
        set outlinedFocusBackgroundColor(a) {
            this.gm = a;
            this.g("OutlinedFocusBackgroundColor");
        }
        get outlinedFocusTextColor() {
            return this.gn;
        }
        set outlinedFocusTextColor(a) {
            this.gn = a;
            this.g("OutlinedFocusTextColor");
        }
        get fabBackgroundColor() {
            return this.e5;
        }
        set fabBackgroundColor(a) {
            this.e5 = a;
            this.g("FabBackgroundColor");
        }
        get fabDisabledBackgroundColor() {
            return this.e7;
        }
        set fabDisabledBackgroundColor(a) {
            this.e7 = a;
            this.g("FabDisabledBackgroundColor");
        }
        get fabDisabledBorderColor() {
            return this.e8;
        }
        set fabDisabledBorderColor(a) {
            this.e8 = a;
            this.g("FabDisabledBorderColor");
        }
        get fabDisabledTextColor() {
            return this.e9;
        }
        set fabDisabledTextColor(a) {
            this.e9 = a;
            this.g("FabDisabledTextColor");
        }
        get fabDisabledElevation() {
            return this.cu;
        }
        set fabDisabledElevation(a) {
            this.cu = a;
            this.g("FabDisabledElevation");
        }
        get fabHoverBackgroundColor() {
            return this.fc;
        }
        set fabHoverBackgroundColor(a) {
            this.fc = a;
            this.g("FabHoverBackgroundColor");
        }
        get fabBorderColor() {
            return this.e6;
        }
        set fabBorderColor(a) {
            this.e6 = a;
            this.g("FabBorderColor");
        }
        get fabRestingElevation() {
            return this.cx;
        }
        set fabRestingElevation(a) {
            this.cx = a;
            this.g("FabRestingElevation");
        }
        get fabHoverElevation() {
            return this.cw;
        }
        set fabHoverElevation(a) {
            this.cw = a;
            this.g("FabHoverElevation");
        }
        get fabFocusElevation() {
            return this.cv;
        }
        set fabFocusElevation(a) {
            this.cv = a;
            this.g("FabFocusElevation");
        }
        get fabBorderWidth() {
            return this.ar;
        }
        set fabBorderWidth(a) {
            this.ar = a;
            this.g("FabBorderWidth");
        }
        get fabCornerRadiusBottomRight() {
            return this.at;
        }
        set fabCornerRadiusBottomRight(a) {
            this.at = a;
            this.g("FabCornerRadiusBottomRight");
        }
        get fabCornerRadiusBottomLeft() {
            return this.as;
        }
        set fabCornerRadiusBottomLeft(a) {
            this.as = a;
            this.g("FabCornerRadiusBottomLeft");
        }
        get fabCornerRadiusTopLeft() {
            return this.au;
        }
        set fabCornerRadiusTopLeft(a) {
            this.au = a;
            this.g("FabCornerRadiusTopLeft");
        }
        get fabCornerRadiusTopRight() {
            return this.av;
        }
        set fabCornerRadiusTopRight(a) {
            this.av = a;
            this.g("FabCornerRadiusTopRight");
        }
        get fabRippleColor() {
            return this.fe;
        }
        set fabRippleColor(a) {
            this.fe = a;
            this.g("FabRippleColor");
        }
        get fabTextColor() {
            return this.ff;
        }
        set fabTextColor(a) {
            this.ff = a;
            this.g("FabTextColor");
        }
        get fabHoverTextColor() {
            return this.fd;
        }
        set fabHoverTextColor(a) {
            this.fd = a;
            this.g("FabHoverTextColor");
        }
        get fabFocusBackgroundColor() {
            return this.fa;
        }
        set fabFocusBackgroundColor(a) {
            this.fa = a;
            this.g("FabFocusBackgroundColor");
        }
        get fabFocusTextColor() {
            return this.fb;
        }
        set fabFocusTextColor(a) {
            this.fb = a;
            this.g("FabFocusTextColor");
        }
        get iconBackgroundColor() {
            return this.f2;
        }
        set iconBackgroundColor(a) {
            this.f2 = a;
            this.g("IconBackgroundColor");
        }
        get iconDisabledBackgroundColor() {
            return this.f4;
        }
        set iconDisabledBackgroundColor(a) {
            this.f4 = a;
            this.g("IconDisabledBackgroundColor");
        }
        get iconDisabledBorderColor() {
            return this.f5;
        }
        set iconDisabledBorderColor(a) {
            this.f5 = a;
            this.g("IconDisabledBorderColor");
        }
        get iconDisabledTextColor() {
            return this.f6;
        }
        set iconDisabledTextColor(a) {
            this.f6 = a;
            this.g("IconDisabledTextColor");
        }
        get iconDisabledElevation() {
            return this.c4;
        }
        set iconDisabledElevation(a) {
            this.c4 = a;
            this.g("IconDisabledElevation");
        }
        get iconHoverBackgroundColor() {
            return this.f9;
        }
        set iconHoverBackgroundColor(a) {
            this.f9 = a;
            this.g("IconHoverBackgroundColor");
        }
        get iconBorderColor() {
            return this.f3;
        }
        set iconBorderColor(a) {
            this.f3 = a;
            this.g("IconBorderColor");
        }
        get iconRestingElevation() {
            return this.c7;
        }
        set iconRestingElevation(a) {
            this.c7 = a;
            this.g("IconRestingElevation");
        }
        get iconHoverElevation() {
            return this.c6;
        }
        set iconHoverElevation(a) {
            this.c6 = a;
            this.g("IconHoverElevation");
        }
        get iconFocusElevation() {
            return this.c5;
        }
        set iconFocusElevation(a) {
            this.c5 = a;
            this.g("IconFocusElevation");
        }
        get iconBorderWidth() {
            return this.a2;
        }
        set iconBorderWidth(a) {
            this.a2 = a;
            this.g("IconBorderWidth");
        }
        get iconCornerRadiusBottomRight() {
            return this.a4;
        }
        set iconCornerRadiusBottomRight(a) {
            this.a4 = a;
            this.g("IconCornerRadiusBottomRight");
        }
        get iconCornerRadiusBottomLeft() {
            return this.a3;
        }
        set iconCornerRadiusBottomLeft(a) {
            this.a3 = a;
            this.g("IconCornerRadiusBottomLeft");
        }
        get iconCornerRadiusTopLeft() {
            return this.a5;
        }
        set iconCornerRadiusTopLeft(a) {
            this.a5 = a;
            this.g("IconCornerRadiusTopLeft");
        }
        get iconCornerRadiusTopRight() {
            return this.a6;
        }
        set iconCornerRadiusTopRight(a) {
            this.a6 = a;
            this.g("IconCornerRadiusTopRight");
        }
        get iconRippleColor() {
            return this.gb;
        }
        set iconRippleColor(a) {
            this.gb = a;
            this.g("IconRippleColor");
        }
        get iconTextColor() {
            return this.gc;
        }
        set iconTextColor(a) {
            this.gc = a;
            this.g("IconTextColor");
        }
        get iconHoverTextColor() {
            return this.ga;
        }
        set iconHoverTextColor(a) {
            this.ga = a;
            this.g("IconHoverTextColor");
        }
        get iconFocusBackgroundColor() {
            return this.f7;
        }
        set iconFocusBackgroundColor(a) {
            this.f7 = a;
            this.g("IconFocusBackgroundColor");
        }
        get iconFocusTextColor() {
            return this.f8;
        }
        set iconFocusTextColor(a) {
            this.f8 = a;
            this.g("IconFocusTextColor");
        }
        get inputId() {
            return this.ge;
        }
        set inputId(a) {
            this.ge = a;
            this.g("InputId");
        }
        get id() {
            return this.gd;
        }
        set id(a) {
            this.gd = a;
            this.g("Id");
        }
        get display() {
            return this.e2;
        }
        set display(a) {
            this.e2 = a;
            this.g("Display");
        }
        get flexDirection() {
            return this.fr;
        }
        set flexDirection(a) {
            this.fr = a;
            this.g("FlexDirection");
        }
        get flexGrow() {
            return this.fs;
        }
        set flexGrow(a) {
            this.fs = a;
            this.g("FlexGrow");
        }
        get alignItems() {
            return this.es;
        }
        set alignItems(a) {
            this.es = a;
            this.g("AlignItems");
        }
        get valueRef() {
            return this.g5;
        }
        set valueRef(a) {
            this.g5 = a;
            this.g("ValueRef");
        }
        get name() {
            return this.gg;
        }
        set name(a) {
            this.gg = a;
            this.g("Name");
        }
        get tabIndex() {
            return this.dh;
        }
        set tabIndex(a) {
            this.dh = a;
            this.g("TabIndex");
        }
        get ariaLabel() {
            return this.et;
        }
        set ariaLabel(a) {
            this.et = a;
            this.g("AriaLabel");
        }
        get fillAvailableSpace() {
            return this.o;
        }
        set fillAvailableSpace(a) {
            this.o = a;
            this.g("FillAvailableSpace");
        }
        get focused() {
            return this.p;
        }
        set focused(a) {
            this.p = a;
            this.g("Focused");
        }
        get isFocusStyleEnabled() {
            return this.q;
        }
        set isFocusStyleEnabled(a) {
            this.q = a;
            this.g("IsFocusStyleEnabled");
        }
        get disableRipple() {
            return this.m;
        }
        set disableRipple(a) {
            this.m = a;
            this.g("DisableRipple");
        }
        get actualDisableRipple() {
            return this.h;
        }
        set actualDisableRipple(a) {
            this.h = a;
            this.g("ActualDisableRipple");
        }
        get disableHover() {
            return this.k;
        }
        set disableHover(a) {
            this.k = a;
            this.g("DisableHover");
        }
        get isHover() {
            return this.r;
        }
        set isHover(a) {
            this.r = a;
            this.g("IsHover");
        }
        get disabled() {
            return this.j;
        }
        set disabled(a) {
            this.j = a;
            this.g("Disabled");
        }
        get disableTransitions() {
            return this.n;
        }
        set disableTransitions(a) {
            this.n = a;
            this.g("DisableTransitions");
        }
        get disablePointer() {
            return this.l;
        }
        set disablePointer(a) {
            this.l = a;
            this.g("DisablePointer");
        }
        get contentPaddingLeft() {
            return this.ak;
        }
        set contentPaddingLeft(a) {
            this.ak = a;
            this.g("ContentPaddingLeft");
        }
        get contentPaddingTop() {
            return this.am;
        }
        set contentPaddingTop(a) {
            this.am = a;
            this.g("ContentPaddingTop");
        }
        get contentPaddingRight() {
            return this.al;
        }
        set contentPaddingRight(a) {
            this.al = a;
            this.g("ContentPaddingRight");
        }
        get contentPaddingBottom() {
            return this.aj;
        }
        set contentPaddingBottom(a) {
            this.aj = a;
            this.g("ContentPaddingBottom");
        }
        get horizontalContentAlignment() {
            return this.fz;
        }
        set horizontalContentAlignment(a) {
            this.fz = a;
            this.g("HorizontalContentAlignment");
        }
        get verticalContentAlignment() {
            return this.g6;
        }
        set verticalContentAlignment(a) {
            this.g6 = a;
            this.g("VerticalContentAlignment");
        }
        get clickTunneling() {
            return this.i;
        }
        set clickTunneling(a) {
            this.i = a;
            this.g("ClickTunneling");
        }
        get clickedRef() {
            return this.ex;
        }
        set clickedRef(a) {
            this.ex = a;
            this.g("ClickedRef");
        }
        get gotFocusRef() {
            return this.fy;
        }
        set gotFocusRef(a) {
            this.fy = a;
            this.g("GotFocusRef");
        }
        get lostFocusRef() {
            return this.gf;
        }
        set lostFocusRef(a) {
            this.gf = a;
            this.g("LostFocusRef");
        }
    }
    ButtonDescription.$t = /*@__PURE__*/ markType(ButtonDescription, 'ButtonDescription', Description.$);
    return ButtonDescription;
})();
