// igniteui-react-core/esm2015/lib/ButtonGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ButtonGroupDescription = /*@__PURE__*/ (() => {
    class ButtonGroupDescription extends Description {
        get_type() {
            return "ButtonGroup";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.au = null;
            this.as = null;
            this.at = null;
            this.af = null;
            this.i = null;
            this.q = 0;
            this.ag = null;
            this.ai = null;
            this.an = null;
            this.ak = null;
            this.ah = null;
            this.aj = null;
            this.p = 0;
            this.ao = null;
            this.ar = null;
            this.al = null;
            this.am = null;
            this.ap = null;
            this.aq = null;
            this.u = 0;
            this.a8 = null;
            this.ba = null;
            this.bf = null;
            this.bc = null;
            this.bd = null;
            this.a9 = null;
            this.bb = null;
            this.t = 0;
            this.be = null;
            this.bt = null;
            this.bw = null;
            this.bu = null;
            this.bv = null;
            this.w = 0;
            this.bh = null;
            this.bj = null;
            this.bo = null;
            this.bl = null;
            this.bm = null;
            this.bi = null;
            this.bk = null;
            this.v = 0;
            this.bn = null;
            this.bp = null;
            this.bs = null;
            this.bq = null;
            this.br = null;
            this.s = 0;
            this.av = null;
            this.ax = null;
            this.a2 = null;
            this.az = null;
            this.a0 = null;
            this.aw = null;
            this.ay = null;
            this.r = 0;
            this.a1 = null;
            this.a3 = null;
            this.a6 = null;
            this.a4 = null;
            this.a5 = null;
            this.bg = null;
            this.a7 = null;
            this.l = false;
            this.m = false;
            this.h = null;
            this.bx = null;
        }
        get displayType() {
            return this.au;
        }
        set displayType(a) {
            this.au = a;
            this.g("DisplayType");
        }
        get baseTheme() {
            return this.as;
        }
        set baseTheme(a) {
            this.as = a;
            this.g("BaseTheme");
        }
        get density() {
            return this.at;
        }
        set density(a) {
            this.at = a;
            this.g("Density");
        }
        get actualDensity() {
            return this.af;
        }
        set actualDensity(a) {
            this.af = a;
            this.g("ActualDensity");
        }
        get selectedIndices() {
            return this.i;
        }
        set selectedIndices(a) {
            this.i = a;
            this.g("SelectedIndices");
        }
        get actualItemCornerRadius() {
            return this.q;
        }
        set actualItemCornerRadius(a) {
            this.q = a;
            this.g("ActualItemCornerRadius");
        }
        get actualItemBackgroundColor() {
            return this.ag;
        }
        set actualItemBackgroundColor(a) {
            this.ag = a;
            this.g("ActualItemBackgroundColor");
        }
        get actualItemDisabledBackgroundColor() {
            return this.ai;
        }
        set actualItemDisabledBackgroundColor(a) {
            this.ai = a;
            this.g("ActualItemDisabledBackgroundColor");
        }
        get actualItemTextColor() {
            return this.an;
        }
        set actualItemTextColor(a) {
            this.an = a;
            this.g("ActualItemTextColor");
        }
        get actualItemDisabledTextColor() {
            return this.ak;
        }
        set actualItemDisabledTextColor(a) {
            this.ak = a;
            this.g("ActualItemDisabledTextColor");
        }
        get actualItemBorderColor() {
            return this.ah;
        }
        set actualItemBorderColor(a) {
            this.ah = a;
            this.g("ActualItemBorderColor");
        }
        get actualItemDisabledBorderColor() {
            return this.aj;
        }
        set actualItemDisabledBorderColor(a) {
            this.aj = a;
            this.g("ActualItemDisabledBorderColor");
        }
        get actualItemBorderWidth() {
            return this.p;
        }
        set actualItemBorderWidth(a) {
            this.p = a;
            this.g("ActualItemBorderWidth");
        }
        get actualSelectedItemBackgroundColor() {
            return this.ao;
        }
        set actualSelectedItemBackgroundColor(a) {
            this.ao = a;
            this.g("ActualSelectedItemBackgroundColor");
        }
        get actualSelectedItemTextColor() {
            return this.ar;
        }
        set actualSelectedItemTextColor(a) {
            this.ar = a;
            this.g("ActualSelectedItemTextColor");
        }
        get actualItemHoverBackgroundColor() {
            return this.al;
        }
        set actualItemHoverBackgroundColor(a) {
            this.al = a;
            this.g("ActualItemHoverBackgroundColor");
        }
        get actualItemHoverTextColor() {
            return this.am;
        }
        set actualItemHoverTextColor(a) {
            this.am = a;
            this.g("ActualItemHoverTextColor");
        }
        get actualSelectedItemHoverBackgroundColor() {
            return this.ap;
        }
        set actualSelectedItemHoverBackgroundColor(a) {
            this.ap = a;
            this.g("ActualSelectedItemHoverBackgroundColor");
        }
        get actualSelectedItemHoverTextColor() {
            return this.aq;
        }
        set actualSelectedItemHoverTextColor(a) {
            this.aq = a;
            this.g("ActualSelectedItemHoverTextColor");
        }
        get itemCornerRadius() {
            return this.u;
        }
        set itemCornerRadius(a) {
            this.u = a;
            this.g("ItemCornerRadius");
        }
        get itemBackgroundColor() {
            return this.a8;
        }
        set itemBackgroundColor(a) {
            this.a8 = a;
            this.g("ItemBackgroundColor");
        }
        get itemDisabledBackgroundColor() {
            return this.ba;
        }
        set itemDisabledBackgroundColor(a) {
            this.ba = a;
            this.g("ItemDisabledBackgroundColor");
        }
        get itemTextColor() {
            return this.bf;
        }
        set itemTextColor(a) {
            this.bf = a;
            this.g("ItemTextColor");
        }
        get itemDisabledTextColor() {
            return this.bc;
        }
        set itemDisabledTextColor(a) {
            this.bc = a;
            this.g("ItemDisabledTextColor");
        }
        get itemHoverBackgroundColor() {
            return this.bd;
        }
        set itemHoverBackgroundColor(a) {
            this.bd = a;
            this.g("ItemHoverBackgroundColor");
        }
        get itemBorderColor() {
            return this.a9;
        }
        set itemBorderColor(a) {
            this.a9 = a;
            this.g("ItemBorderColor");
        }
        get itemDisabledBorderColor() {
            return this.bb;
        }
        set itemDisabledBorderColor(a) {
            this.bb = a;
            this.g("ItemDisabledBorderColor");
        }
        get itemBorderWidth() {
            return this.t;
        }
        set itemBorderWidth(a) {
            this.t = a;
            this.g("ItemBorderWidth");
        }
        get itemHoverTextColor() {
            return this.be;
        }
        set itemHoverTextColor(a) {
            this.be = a;
            this.g("ItemHoverTextColor");
        }
        get selectedItemBackgroundColor() {
            return this.bt;
        }
        set selectedItemBackgroundColor(a) {
            this.bt = a;
            this.g("SelectedItemBackgroundColor");
        }
        get selectedItemTextColor() {
            return this.bw;
        }
        set selectedItemTextColor(a) {
            this.bw = a;
            this.g("SelectedItemTextColor");
        }
        get selectedItemHoverBackgroundColor() {
            return this.bu;
        }
        set selectedItemHoverBackgroundColor(a) {
            this.bu = a;
            this.g("SelectedItemHoverBackgroundColor");
        }
        get selectedItemHoverTextColor() {
            return this.bv;
        }
        set selectedItemHoverTextColor(a) {
            this.bv = a;
            this.g("SelectedItemHoverTextColor");
        }
        get outlinedItemCornerRadius() {
            return this.w;
        }
        set outlinedItemCornerRadius(a) {
            this.w = a;
            this.g("OutlinedItemCornerRadius");
        }
        get outlinedItemBackgroundColor() {
            return this.bh;
        }
        set outlinedItemBackgroundColor(a) {
            this.bh = a;
            this.g("OutlinedItemBackgroundColor");
        }
        get outlinedItemDisabledBackgroundColor() {
            return this.bj;
        }
        set outlinedItemDisabledBackgroundColor(a) {
            this.bj = a;
            this.g("OutlinedItemDisabledBackgroundColor");
        }
        get outlinedItemTextColor() {
            return this.bo;
        }
        set outlinedItemTextColor(a) {
            this.bo = a;
            this.g("OutlinedItemTextColor");
        }
        get outlinedItemDisabledTextColor() {
            return this.bl;
        }
        set outlinedItemDisabledTextColor(a) {
            this.bl = a;
            this.g("OutlinedItemDisabledTextColor");
        }
        get outlinedItemHoverBackgroundColor() {
            return this.bm;
        }
        set outlinedItemHoverBackgroundColor(a) {
            this.bm = a;
            this.g("OutlinedItemHoverBackgroundColor");
        }
        get outlinedItemBorderColor() {
            return this.bi;
        }
        set outlinedItemBorderColor(a) {
            this.bi = a;
            this.g("OutlinedItemBorderColor");
        }
        get outlinedItemDisabledBorderColor() {
            return this.bk;
        }
        set outlinedItemDisabledBorderColor(a) {
            this.bk = a;
            this.g("OutlinedItemDisabledBorderColor");
        }
        get outlinedItemBorderWidth() {
            return this.v;
        }
        set outlinedItemBorderWidth(a) {
            this.v = a;
            this.g("OutlinedItemBorderWidth");
        }
        get outlinedItemHoverTextColor() {
            return this.bn;
        }
        set outlinedItemHoverTextColor(a) {
            this.bn = a;
            this.g("OutlinedItemHoverTextColor");
        }
        get outlinedSelectedItemBackgroundColor() {
            return this.bp;
        }
        set outlinedSelectedItemBackgroundColor(a) {
            this.bp = a;
            this.g("OutlinedSelectedItemBackgroundColor");
        }
        get outlinedSelectedItemTextColor() {
            return this.bs;
        }
        set outlinedSelectedItemTextColor(a) {
            this.bs = a;
            this.g("OutlinedSelectedItemTextColor");
        }
        get outlinedSelectedItemHoverBackgroundColor() {
            return this.bq;
        }
        set outlinedSelectedItemHoverBackgroundColor(a) {
            this.bq = a;
            this.g("OutlinedSelectedItemHoverBackgroundColor");
        }
        get outlinedSelectedItemHoverTextColor() {
            return this.br;
        }
        set outlinedSelectedItemHoverTextColor(a) {
            this.br = a;
            this.g("OutlinedSelectedItemHoverTextColor");
        }
        get flatItemCornerRadius() {
            return this.s;
        }
        set flatItemCornerRadius(a) {
            this.s = a;
            this.g("FlatItemCornerRadius");
        }
        get flatItemBackgroundColor() {
            return this.av;
        }
        set flatItemBackgroundColor(a) {
            this.av = a;
            this.g("FlatItemBackgroundColor");
        }
        get flatItemDisabledBackgroundColor() {
            return this.ax;
        }
        set flatItemDisabledBackgroundColor(a) {
            this.ax = a;
            this.g("FlatItemDisabledBackgroundColor");
        }
        get flatItemTextColor() {
            return this.a2;
        }
        set flatItemTextColor(a) {
            this.a2 = a;
            this.g("FlatItemTextColor");
        }
        get flatItemDisabledTextColor() {
            return this.az;
        }
        set flatItemDisabledTextColor(a) {
            this.az = a;
            this.g("FlatItemDisabledTextColor");
        }
        get flatItemHoverBackgroundColor() {
            return this.a0;
        }
        set flatItemHoverBackgroundColor(a) {
            this.a0 = a;
            this.g("FlatItemHoverBackgroundColor");
        }
        get flatItemBorderColor() {
            return this.aw;
        }
        set flatItemBorderColor(a) {
            this.aw = a;
            this.g("FlatItemBorderColor");
        }
        get flatItemDisabledBorderColor() {
            return this.ay;
        }
        set flatItemDisabledBorderColor(a) {
            this.ay = a;
            this.g("FlatItemDisabledBorderColor");
        }
        get flatItemBorderWidth() {
            return this.r;
        }
        set flatItemBorderWidth(a) {
            this.r = a;
            this.g("FlatItemBorderWidth");
        }
        get flatItemHoverTextColor() {
            return this.a1;
        }
        set flatItemHoverTextColor(a) {
            this.a1 = a;
            this.g("FlatItemHoverTextColor");
        }
        get flatSelectedItemBackgroundColor() {
            return this.a3;
        }
        set flatSelectedItemBackgroundColor(a) {
            this.a3 = a;
            this.g("FlatSelectedItemBackgroundColor");
        }
        get flatSelectedItemTextColor() {
            return this.a6;
        }
        set flatSelectedItemTextColor(a) {
            this.a6 = a;
            this.g("FlatSelectedItemTextColor");
        }
        get flatSelectedItemHoverBackgroundColor() {
            return this.a4;
        }
        set flatSelectedItemHoverBackgroundColor(a) {
            this.a4 = a;
            this.g("FlatSelectedItemHoverBackgroundColor");
        }
        get flatSelectedItemHoverTextColor() {
            return this.a5;
        }
        set flatSelectedItemHoverTextColor(a) {
            this.a5 = a;
            this.g("FlatSelectedItemHoverTextColor");
        }
        get orientation() {
            return this.bg;
        }
        set orientation(a) {
            this.bg = a;
            this.g("Orientation");
        }
        get id() {
            return this.a7;
        }
        set id(a) {
            this.a7 = a;
            this.g("Id");
        }
        get disabled() {
            return this.l;
        }
        set disabled(a) {
            this.l = a;
            this.g("Disabled");
        }
        get isMultiSelect() {
            return this.m;
        }
        set isMultiSelect(a) {
            this.m = a;
            this.g("IsMultiSelect");
        }
        get buttons() {
            return this.h;
        }
        set buttons(a) {
            this.h = a;
            this.g("Buttons");
        }
        get selectionChangedRef() {
            return this.bx;
        }
        set selectionChangedRef(a) {
            this.bx = a;
            this.g("SelectionChangedRef");
        }
    }
    ButtonGroupDescription.$t = /*@__PURE__*/ markType(ButtonGroupDescription, 'ButtonGroupDescription', Description.$);
    return ButtonGroupDescription;
})();
