// igniteui-react-core/esm2015/lib/CalendarDateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CalendarDateDescription = /*@__PURE__*/ (() => {
    class CalendarDateDescription extends Description {
        get_type() {
            return "CalendarDate";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.o = new Date();
            this.i = false;
            this.k = false;
            this.j = false;
        }
        get date() {
            return this.o;
        }
        set date(a) {
            this.o = a;
            this.g("Date");
        }
        get isCurrentMonth() {
            return this.i;
        }
        set isCurrentMonth(a) {
            this.i = a;
            this.g("IsCurrentMonth");
        }
        get isPrevMonth() {
            return this.k;
        }
        set isPrevMonth(a) {
            this.k = a;
            this.g("IsPrevMonth");
        }
        get isNextMonth() {
            return this.j;
        }
        set isNextMonth(a) {
            this.j = a;
            this.g("IsNextMonth");
        }
    }
    CalendarDateDescription.$t = /*@__PURE__*/ markType(CalendarDateDescription, 'CalendarDateDescription', Description.$);
    CalendarDateDescription.__marshalByValue = true;
    CalendarDateDescription.__marshalByValueAlias = "CalendarDate";
    return CalendarDateDescription;
})();
