// igniteui-react-core/esm2015/lib/CalendarFormatOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CalendarFormatOptionsDescription = /*@__PURE__*/ (() => {
    class CalendarFormatOptionsDescription extends Description {
        get_type() {
            return "CalendarFormatOptions";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = null;
            this.j = null;
        }
        get weekday() {
            return this.k;
        }
        set weekday(a) {
            this.k = a;
            this.g("Weekday");
        }
        get month() {
            return this.j;
        }
        set month(a) {
            this.j = a;
            this.g("Month");
        }
    }
    CalendarFormatOptionsDescription.$t = /*@__PURE__*/ markType(CalendarFormatOptionsDescription, 'CalendarFormatOptionsDescription', Description.$);
    CalendarFormatOptionsDescription.__marshalByValue = true;
    CalendarFormatOptionsDescription.__marshalByValueAlias = "CalendarFormatOptions";
    return CalendarFormatOptionsDescription;
})();
