// igniteui-react-core/esm2015/lib/CalloutLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CalloutLayerDescription = /*@__PURE__*/ (() => {
    class CalloutLayerDescription extends AnnotationLayerDescription {
        get_type() {
            return "CalloutLayer";
        }
        constructor() {
            super();
            this.gd = false;
            this.g2 = 0;
            this.g0 = 0;
            this.g3 = 0;
            this.g1 = 0;
            this.h7 = null;
            this.ib = null;
            this.ic = null;
            this.h5 = null;
            this.gf = false;
            this.h6 = null;
            this.ga = null;
            this.h3 = null;
            this.h8 = null;
            this.h9 = null;
            this.h2 = null;
            this.f9 = null;
            this.ge = false;
            this.gj = false;
            this.gm = false;
            this.gi = false;
            this.gk = false;
            this.gl = false;
            this.gn = false;
            this.h4 = null;
            this.hp = null;
            this.gz = 0;
            this.gh = false;
            this.gg = false;
            this.hl = 0;
            this.h1 = null;
            this.hw = null;
            this.ht = null;
            this.hq = null;
            this.hm = 0;
            this.hx = null;
            this.hv = null;
            this.g5 = 0;
            this.g7 = 0;
            this.g6 = 0;
            this.g4 = 0;
            this.hr = null;
            this.g8 = 0;
            this.g9 = 0;
            this.ia = null;
            this.h0 = null;
            this.hy = null;
            this.hs = null;
            this.hu = null;
            this.hz = null;
        }
        get calloutExpandsAxisBufferEnabled() {
            return this.gd;
        }
        set calloutExpandsAxisBufferEnabled(a) {
            this.gd = a;
            this.g("CalloutExpandsAxisBufferEnabled");
        }
        get calloutExpandsAxisBufferMinHeight() {
            return this.g2;
        }
        set calloutExpandsAxisBufferMinHeight(a) {
            this.g2 = a;
            this.g("CalloutExpandsAxisBufferMinHeight");
        }
        get calloutExpandsAxisBufferMaxHeight() {
            return this.g0;
        }
        set calloutExpandsAxisBufferMaxHeight(a) {
            this.g0 = a;
            this.g("CalloutExpandsAxisBufferMaxHeight");
        }
        get calloutExpandsAxisBufferMinWidth() {
            return this.g3;
        }
        set calloutExpandsAxisBufferMinWidth(a) {
            this.g3 = a;
            this.g("CalloutExpandsAxisBufferMinWidth");
        }
        get calloutExpandsAxisBufferMaxWidth() {
            return this.g1;
        }
        set calloutExpandsAxisBufferMaxWidth(a) {
            this.g1 = a;
            this.g("CalloutExpandsAxisBufferMaxWidth");
        }
        get labelMemberPath() {
            return this.h7;
        }
        set labelMemberPath(a) {
            this.h7 = a;
            this.g("LabelMemberPath");
        }
        get xMemberPath() {
            return this.ib;
        }
        set xMemberPath(a) {
            this.ib = a;
            this.g("XMemberPath");
        }
        get yMemberPath() {
            return this.ic;
        }
        set yMemberPath(a) {
            this.ic = a;
            this.g("YMemberPath");
        }
        get keyMemberPath() {
            return this.h5;
        }
        set keyMemberPath(a) {
            this.h5 = a;
            this.g("KeyMemberPath");
        }
        get isCalloutOffsettingEnabled() {
            return this.gf;
        }
        set isCalloutOffsettingEnabled(a) {
            this.gf = a;
            this.g("IsCalloutOffsettingEnabled");
        }
        get labelFormat() {
            return this.h6;
        }
        set labelFormat(a) {
            this.h6 = a;
            this.g("LabelFormat");
        }
        get labelFormatSpecifiers() {
            return this.ga;
        }
        set labelFormatSpecifiers(a) {
            this.ga = a;
            this.g("LabelFormatSpecifiers");
        }
        get contentMemberPath() {
            return this.h3;
        }
        set contentMemberPath(a) {
            this.h3 = a;
            this.g("ContentMemberPath");
        }
        get targetSeriesName() {
            return this.h8;
        }
        set targetSeriesName(a) {
            this.h8 = a;
            this.g("TargetSeriesName");
        }
        get targetSeriesRef() {
            return this.h9;
        }
        set targetSeriesRef(a) {
            this.h9 = a;
            this.g("TargetSeriesRef");
        }
        get collisionChannel() {
            return this.h2;
        }
        set collisionChannel(a) {
            this.h2 = a;
            this.g("CollisionChannel");
        }
        get allowedPositions() {
            return this.f9;
        }
        set allowedPositions(a) {
            this.f9 = a;
            this.g("AllowedPositions");
        }
        get isAutoCalloutBehaviorEnabled() {
            return this.ge;
        }
        set isAutoCalloutBehaviorEnabled(a) {
            this.ge = a;
            this.g("IsAutoCalloutBehaviorEnabled");
        }
        get useInterpolatedValueForAutoCalloutLabels() {
            return this.gj;
        }
        set useInterpolatedValueForAutoCalloutLabels(a) {
            this.gj = a;
            this.g("UseInterpolatedValueForAutoCalloutLabels");
        }
        get useSeriesColorForOutline() {
            return this.gm;
        }
        set useSeriesColorForOutline(a) {
            this.gm = a;
            this.g("UseSeriesColorForOutline");
        }
        get useAutoContrastingLabelColors() {
            return this.gi;
        }
        set useAutoContrastingLabelColors(a) {
            this.gi = a;
            this.g("UseAutoContrastingLabelColors");
        }
        get useItemColorForFill() {
            return this.gk;
        }
        set useItemColorForFill(a) {
            this.gk = a;
            this.g("UseItemColorForFill");
        }
        get useItemColorForOutline() {
            return this.gl;
        }
        set useItemColorForOutline(a) {
            this.gl = a;
            this.g("UseItemColorForOutline");
        }
        get useValueForAutoCalloutLabels() {
            return this.gn;
        }
        set useValueForAutoCalloutLabels(a) {
            this.gn = a;
            this.g("UseValueForAutoCalloutLabels");
        }
        get highlightedValueLabelMode() {
            return this.h4;
        }
        set highlightedValueLabelMode(a) {
            this.h4 = a;
            this.g("HighlightedValueLabelMode");
        }
        get autoCalloutVisibilityMode() {
            return this.hp;
        }
        set autoCalloutVisibilityMode(a) {
            this.hp = a;
            this.g("AutoCalloutVisibilityMode");
        }
        get calloutCornerRadius() {
            return this.gz;
        }
        set calloutCornerRadius(a) {
            this.gz = a;
            this.g("CalloutCornerRadius");
        }
        get isCustomCalloutStyleEnabled() {
            return this.gh;
        }
        set isCustomCalloutStyleEnabled(a) {
            this.gh = a;
            this.g("IsCustomCalloutStyleEnabled");
        }
        get isCustomCalloutRenderStyleEnabled() {
            return this.gg;
        }
        set isCustomCalloutRenderStyleEnabled(a) {
            this.gg = a;
            this.g("IsCustomCalloutRenderStyleEnabled");
        }
        get autoCalloutLabelPrecision() {
            return this.hl;
        }
        set autoCalloutLabelPrecision(a) {
            this.hl = a;
            this.g("AutoCalloutLabelPrecision");
        }
        get calloutTextColor() {
            return this.h1;
        }
        set calloutTextColor(a) {
            this.h1 = a;
            this.g("CalloutTextColor");
        }
        get calloutLightTextColor() {
            return this.hw;
        }
        set calloutLightTextColor(a) {
            this.hw = a;
            this.g("CalloutLightTextColor");
        }
        get calloutDarkTextColor() {
            return this.ht;
        }
        set calloutDarkTextColor(a) {
            this.ht = a;
            this.g("CalloutDarkTextColor");
        }
        get calloutBackground() {
            return this.hq;
        }
        set calloutBackground(a) {
            this.hq = a;
            this.g("CalloutBackground");
        }
        get calloutInterpolatedValuePrecision() {
            return this.hm;
        }
        set calloutInterpolatedValuePrecision(a) {
            this.hm = a;
            this.g("CalloutInterpolatedValuePrecision");
        }
        get calloutOutline() {
            return this.hx;
        }
        set calloutOutline(a) {
            this.hx = a;
            this.g("CalloutOutline");
        }
        get calloutLeaderBrush() {
            return this.hv;
        }
        set calloutLeaderBrush(a) {
            this.hv = a;
            this.g("CalloutLeaderBrush");
        }
        get calloutPaddingLeft() {
            return this.g5;
        }
        set calloutPaddingLeft(a) {
            this.g5 = a;
            this.g("CalloutPaddingLeft");
        }
        get calloutPaddingTop() {
            return this.g7;
        }
        set calloutPaddingTop(a) {
            this.g7 = a;
            this.g("CalloutPaddingTop");
        }
        get calloutPaddingRight() {
            return this.g6;
        }
        set calloutPaddingRight(a) {
            this.g6 = a;
            this.g("CalloutPaddingRight");
        }
        get calloutPaddingBottom() {
            return this.g4;
        }
        set calloutPaddingBottom(a) {
            this.g4 = a;
            this.g("CalloutPaddingBottom");
        }
        get calloutCollisionMode() {
            return this.hr;
        }
        set calloutCollisionMode(a) {
            this.hr = a;
            this.g("CalloutCollisionMode");
        }
        get calloutPositionPadding() {
            return this.g8;
        }
        set calloutPositionPadding(a) {
            this.g8 = a;
            this.g("CalloutPositionPadding");
        }
        get calloutStrokeThickness() {
            return this.g9;
        }
        set calloutStrokeThickness(a) {
            this.g9 = a;
            this.g("CalloutStrokeThickness");
        }
        get textStyle() {
            return this.ia;
        }
        set textStyle(a) {
            this.ia = a;
            this.g("TextStyle");
        }
        get calloutStyleUpdatingRef() {
            return this.h0;
        }
        set calloutStyleUpdatingRef(a) {
            this.h0 = a;
            this.g("CalloutStyleUpdatingRef");
        }
        get calloutRenderStyleUpdatingRef() {
            return this.hy;
        }
        set calloutRenderStyleUpdatingRef(a) {
            this.hy = a;
            this.g("CalloutRenderStyleUpdatingRef");
        }
        get calloutContentUpdatingRef() {
            return this.hs;
        }
        set calloutContentUpdatingRef(a) {
            this.hs = a;
            this.g("CalloutContentUpdatingRef");
        }
        get calloutLabelUpdatingRef() {
            return this.hu;
        }
        set calloutLabelUpdatingRef(a) {
            this.hu = a;
            this.g("CalloutLabelUpdatingRef");
        }
        get calloutSeriesSelectingRef() {
            return this.hz;
        }
        set calloutSeriesSelectingRef(a) {
            this.hz = a;
            this.g("CalloutSeriesSelectingRef");
        }
    }
    CalloutLayerDescription.$t = /*@__PURE__*/ markType(CalloutLayerDescription, 'CalloutLayerDescription', AnnotationLayerDescription.$);
    return CalloutLayerDescription;
})();
