// igniteui-react-core/esm2015/lib/CalloutStyleUpdatingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CalloutStyleUpdatingEventArgsDescription = /*@__PURE__*/ (() => {
    class CalloutStyleUpdatingEventArgsDescription extends Description {
        get_type() {
            return "CalloutStyleUpdatingEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.q = null;
            this.r = null;
            this.m = null;
            this.h = null;
            this.l = null;
            this.p = null;
            this.o = null;
            this.n = null;
            this.j = 0;
        }
        get xValueRef() {
            return this.q;
        }
        set xValueRef(a) {
            this.q = a;
            this.g("XValueRef");
        }
        get yValueRef() {
            return this.r;
        }
        set yValueRef(a) {
            this.r = a;
            this.g("YValueRef");
        }
        get itemRef() {
            return this.m;
        }
        set itemRef(a) {
            this.m = a;
            this.g("ItemRef");
        }
        get series() {
            return this.h;
        }
        set series(a) {
            this.h = a;
            this.g("Series");
        }
        get background() {
            return this.l;
        }
        set background(a) {
            this.l = a;
            this.g("Background");
        }
        get textColor() {
            return this.p;
        }
        set textColor(a) {
            this.p = a;
            this.g("TextColor");
        }
        get outline() {
            return this.o;
        }
        set outline(a) {
            this.o = a;
            this.g("Outline");
        }
        get leaderBrush() {
            return this.n;
        }
        set leaderBrush(a) {
            this.n = a;
            this.g("LeaderBrush");
        }
        get strokeThickness() {
            return this.j;
        }
        set strokeThickness(a) {
            this.j = a;
            this.g("StrokeThickness");
        }
    }
    CalloutStyleUpdatingEventArgsDescription.$t = /*@__PURE__*/ markType(CalloutStyleUpdatingEventArgsDescription, 'CalloutStyleUpdatingEventArgsDescription', Description.$);
    return CalloutStyleUpdatingEventArgsDescription;
})();
