// igniteui-react-core/esm2015/lib/CalloutLabelUpdatingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CalloutLabelUpdatingEventArgsDescription = /*@__PURE__*/ (() => {
    class CalloutLabelUpdatingEventArgsDescription extends Description {
        get_type() {
            return "CalloutLabelUpdatingEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.m = null;
            this.n = null;
            this.j = null;
            this.h = null;
            this.l = null;
            this.k = null;
        }
        get xValueRef() {
            return this.m;
        }
        set xValueRef(a) {
            this.m = a;
            this.g("XValueRef");
        }
        get yValueRef() {
            return this.n;
        }
        set yValueRef(a) {
            this.n = a;
            this.g("YValueRef");
        }
        get itemRef() {
            return this.j;
        }
        set itemRef(a) {
            this.j = a;
            this.g("ItemRef");
        }
        get series() {
            return this.h;
        }
        set series(a) {
            this.h = a;
            this.g("Series");
        }
        get seriesName() {
            return this.l;
        }
        set seriesName(a) {
            this.l = a;
            this.g("SeriesName");
        }
        get labelRef() {
            return this.k;
        }
        set labelRef(a) {
            this.k = a;
            this.g("LabelRef");
        }
    }
    CalloutLabelUpdatingEventArgsDescription.$t = /*@__PURE__*/ markType(CalloutLabelUpdatingEventArgsDescription, 'CalloutLabelUpdatingEventArgsDescription', Description.$);
    return CalloutLabelUpdatingEventArgsDescription;
})();
