// igniteui-react-core/esm2015/lib/CalloutCollisionMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * An enum representing the available collision modes of callout layers.
 */
export var CalloutCollisionMode = /*@__PURE__*/ (function (CalloutCollisionMode) {
    /**
     * Automatically decide the collision strategy.
     */
    CalloutCollisionMode[CalloutCollisionMode["Auto"] = 0] = "Auto";
    /**
     * Use a simulated annealing based collision strategy. This is higher quality, but takes longer, and is performed time-sliced in the background until an acceptable quality is reached.
     */
    CalloutCollisionMode[CalloutCollisionMode["SimulatedAnnealing"] = 1] = "SimulatedAnnealing";
    /**
     * Use a greedy algorithm to avoid collisions. This is cheap and predictable, but of comparatively low quality.
     */
    CalloutCollisionMode[CalloutCollisionMode["Greedy"] = 2] = "Greedy";
    /**
     * Use a greedy algorithm with localized center of mass hints to avoid collisions. This is relatively cheap to perform, compared to the simulated annealing approach, but is of comparatively lower quality.
     */
    CalloutCollisionMode[CalloutCollisionMode["GreedyCenterOfMass"] = 3] = "GreedyCenterOfMass";
    /**
     * If slice anchors are present, either fit the labels inside the slices or outside, depending on collisions.
     */
    CalloutCollisionMode[CalloutCollisionMode["RadialBestFit"] = 4] = "RadialBestFit";
    /**
     * If slice anchors are present, either fit the labels inside the slices or outside, depending on collisions.
     */
    CalloutCollisionMode[CalloutCollisionMode["RadialInsideEnd"] = 5] = "RadialInsideEnd";
    /**
     * If slice anchors are present, either fit the labels inside the slices or outside, depending on collisions.
     */
    CalloutCollisionMode[CalloutCollisionMode["RadialOutsideEnd"] = 6] = "RadialOutsideEnd";
    /**
     * If slice anchors are present, fit the labels inside the slices.
     */
    CalloutCollisionMode[CalloutCollisionMode["RadialCenter"] = 7] = "RadialCenter";
    return CalloutCollisionMode;
})({});
/**
 * @hidden
 */
export let CalloutCollisionMode_$type = /*@__PURE__*/ markEnum('CalloutCollisionMode', 'Auto,0|SimulatedAnnealing,1|Greedy,2|GreedyCenterOfMass,3|RadialBestFit,4|RadialInsideEnd,5|RadialOutsideEnd,6|RadialCenter,7');
