// igniteui-react-core/esm2015/lib/CalloutRenderStyleUpdatingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CalloutRenderStyleUpdatingEventArgsDescription = /*@__PURE__*/ (() => {
    class CalloutRenderStyleUpdatingEventArgsDescription extends Description {
        get_type() {
            return "CalloutRenderStyleUpdatingEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.z = null;
            this.aa = null;
            this.j = 0;
            this.k = 0;
            this.m = 0;
            this.n = 0;
            this.t = null;
            this.v = null;
            this.h = null;
            this.u = null;
            this.y = null;
            this.x = null;
            this.w = null;
            this.l = 0;
        }
        get xValueRef() {
            return this.z;
        }
        set xValueRef(a) {
            this.z = a;
            this.g("XValueRef");
        }
        get yValueRef() {
            return this.aa;
        }
        set yValueRef(a) {
            this.aa = a;
            this.g("YValueRef");
        }
        get labelPositionX() {
            return this.j;
        }
        set labelPositionX(a) {
            this.j = a;
            this.g("LabelPositionX");
        }
        get labelPositionY() {
            return this.k;
        }
        set labelPositionY(a) {
            this.k = a;
            this.g("LabelPositionY");
        }
        get targetPositionX() {
            return this.m;
        }
        set targetPositionX(a) {
            this.m = a;
            this.g("TargetPositionX");
        }
        get targetPositionY() {
            return this.n;
        }
        set targetPositionY(a) {
            this.n = a;
            this.g("TargetPositionY");
        }
        get actualPosition() {
            return this.t;
        }
        set actualPosition(a) {
            this.t = a;
            this.g("ActualPosition");
        }
        get itemRef() {
            return this.v;
        }
        set itemRef(a) {
            this.v = a;
            this.g("ItemRef");
        }
        get series() {
            return this.h;
        }
        set series(a) {
            this.h = a;
            this.g("Series");
        }
        get background() {
            return this.u;
        }
        set background(a) {
            this.u = a;
            this.g("Background");
        }
        get textColor() {
            return this.y;
        }
        set textColor(a) {
            this.y = a;
            this.g("TextColor");
        }
        get outline() {
            return this.x;
        }
        set outline(a) {
            this.x = a;
            this.g("Outline");
        }
        get leaderBrush() {
            return this.w;
        }
        set leaderBrush(a) {
            this.w = a;
            this.g("LeaderBrush");
        }
        get strokeThickness() {
            return this.l;
        }
        set strokeThickness(a) {
            this.l = a;
            this.g("StrokeThickness");
        }
    }
    CalloutRenderStyleUpdatingEventArgsDescription.$t = /*@__PURE__*/ markType(CalloutRenderStyleUpdatingEventArgsDescription, 'CalloutRenderStyleUpdatingEventArgsDescription', Description.$);
    return CalloutRenderStyleUpdatingEventArgsDescription;
})();
