// igniteui-react-core/esm2015/lib/CalloutContentUpdatingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CalloutContentUpdatingEventArgsDescription = /*@__PURE__*/ (() => {
    class CalloutContentUpdatingEventArgsDescription extends Description {
        get_type() {
            return "CalloutContentUpdatingEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.k = null;
            this.i = null;
            this.h = null;
        }
        get xValueRef() {
            return this.j;
        }
        set xValueRef(a) {
            this.j = a;
            this.g("XValueRef");
        }
        get yValueRef() {
            return this.k;
        }
        set yValueRef(a) {
            this.k = a;
            this.g("YValueRef");
        }
        get itemRef() {
            return this.i;
        }
        set itemRef(a) {
            this.i = a;
            this.g("ItemRef");
        }
        get contentRef() {
            return this.h;
        }
        set contentRef(a) {
            this.h = a;
            this.g("ContentRef");
        }
    }
    CalloutContentUpdatingEventArgsDescription.$t = /*@__PURE__*/ markType(CalloutContentUpdatingEventArgsDescription, 'CalloutContentUpdatingEventArgsDescription', Description.$);
    return CalloutContentUpdatingEventArgsDescription;
})();
