// igniteui-react-core/esm2015/lib/CalloutAnnotationDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CalloutAnnotationDescription = /*@__PURE__*/ (() => {
    class CalloutAnnotationDescription extends Description {
        get_type() {
            return "CalloutAnnotation";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.ae = null;
            this.af = null;
            this.ac = null;
            this.z = null;
            this.w = null;
            this.h = null;
            this.y = null;
            this.x = null;
            this.ad = null;
            this.v = null;
            this.ab = null;
            this.aa = null;
            this.o = 0;
            this.l = 0;
            this.n = 0;
            this.m = 0;
            this.k = 0;
            this.j = 0;
        }
        get xValueRef() {
            return this.ae;
        }
        set xValueRef(a) {
            this.ae = a;
            this.g("XValueRef");
        }
        get yValueRef() {
            return this.af;
        }
        set yValueRef(a) {
            this.af = a;
            this.g("YValueRef");
        }
        get text() {
            return this.ac;
        }
        set text(a) {
            this.ac = a;
            this.g("Text");
        }
        get keyRef() {
            return this.z;
        }
        set keyRef(a) {
            this.z = a;
            this.g("KeyRef");
        }
        get contentRef() {
            return this.w;
        }
        set contentRef(a) {
            this.w = a;
            this.g("ContentRef");
        }
        get series() {
            return this.h;
        }
        set series(a) {
            this.h = a;
            this.g("Series");
        }
        get itemColor() {
            return this.y;
        }
        set itemColor(a) {
            this.y = a;
            this.g("ItemColor");
        }
        get formatLabelRef() {
            return this.x;
        }
        set formatLabelRef(a) {
            this.x = a;
            this.g("FormatLabelRef");
        }
        get textColor() {
            return this.ad;
        }
        set textColor(a) {
            this.ad = a;
            this.g("TextColor");
        }
        get background() {
            return this.v;
        }
        set background(a) {
            this.v = a;
            this.g("Background");
        }
        get outline() {
            return this.ab;
        }
        set outline(a) {
            this.ab = a;
            this.g("Outline");
        }
        get leaderBrush() {
            return this.aa;
        }
        set leaderBrush(a) {
            this.aa = a;
            this.g("LeaderBrush");
        }
        get strokeThickness() {
            return this.o;
        }
        set strokeThickness(a) {
            this.o = a;
            this.g("StrokeThickness");
        }
        get backgroundPaddingLeft() {
            return this.l;
        }
        set backgroundPaddingLeft(a) {
            this.l = a;
            this.g("BackgroundPaddingLeft");
        }
        get backgroundPaddingTop() {
            return this.n;
        }
        set backgroundPaddingTop(a) {
            this.n = a;
            this.g("BackgroundPaddingTop");
        }
        get backgroundPaddingRight() {
            return this.m;
        }
        set backgroundPaddingRight(a) {
            this.m = a;
            this.g("BackgroundPaddingRight");
        }
        get backgroundPaddingBottom() {
            return this.k;
        }
        set backgroundPaddingBottom(a) {
            this.k = a;
            this.g("BackgroundPaddingBottom");
        }
        get backgroundCornerRadius() {
            return this.j;
        }
        set backgroundCornerRadius(a) {
            this.j = a;
            this.g("BackgroundCornerRadius");
        }
    }
    CalloutAnnotationDescription.$t = /*@__PURE__*/ markType(CalloutAnnotationDescription, 'CalloutAnnotationDescription', Description.$);
    return CalloutAnnotationDescription;
})();
