// igniteui-react-core/esm2015/lib/CaptureImageSettingsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CaptureImageSettingsDescription = /*@__PURE__*/ (() => {
    class CaptureImageSettingsDescription extends Description {
        get_type() {
            return "CaptureImageSettings";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = false;
        }
        get format() {
            return this.j;
        }
        set format(a) {
            this.j = a;
            this.g("Format");
        }
        get addToClipboard() {
            return this.h;
        }
        set addToClipboard(a) {
            this.h = a;
            this.g("AddToClipboard");
        }
    }
    CaptureImageSettingsDescription.$t = /*@__PURE__*/ markType(CaptureImageSettingsDescription, 'CaptureImageSettingsDescription', Description.$);
    return CaptureImageSettingsDescription;
})();
