// igniteui-react-core/esm2015/lib/CategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CategorySeriesDescription = /*@__PURE__*/ (() => {
    class CategorySeriesDescription extends MarkerSeriesDescription {
        get_type() {
            return "CategorySeries";
        }
        constructor() {
            super();
            this.fs = false;
            this.fr = false;
            this.f1 = null;
            this.fu = false;
            this.f2 = null;
            this.ft = false;
            this.f0 = null;
            this.fz = null;
        }
        get isCustomCategoryStyleAllowed() {
            return this.fs;
        }
        set isCustomCategoryStyleAllowed(a) {
            this.fs = a;
            this.g("IsCustomCategoryStyleAllowed");
        }
        get isCustomCategoryMarkerStyleAllowed() {
            return this.fr;
        }
        set isCustomCategoryMarkerStyleAllowed(a) {
            this.fr = a;
            this.g("IsCustomCategoryMarkerStyleAllowed");
        }
        get categoryCollisionMode() {
            return this.f1;
        }
        set categoryCollisionMode(a) {
            this.f1 = a;
            this.g("CategoryCollisionMode");
        }
        get useHighMarkerFidelity() {
            return this.fu;
        }
        set useHighMarkerFidelity(a) {
            this.fu = a;
            this.g("UseHighMarkerFidelity");
        }
        get transitionInMode() {
            return this.f2;
        }
        set transitionInMode(a) {
            this.f2 = a;
            this.g("TransitionInMode");
        }
        get isTransitionInEnabled() {
            return this.ft;
        }
        set isTransitionInEnabled(a) {
            this.ft = a;
            this.g("IsTransitionInEnabled");
        }
        get assigningCategoryStyleRef() {
            return this.f0;
        }
        set assigningCategoryStyleRef(a) {
            this.f0 = a;
            this.g("AssigningCategoryStyleRef");
        }
        get assigningCategoryMarkerStyleRef() {
            return this.fz;
        }
        set assigningCategoryMarkerStyleRef(a) {
            this.fz = a;
            this.g("AssigningCategoryMarkerStyleRef");
        }
    }
    CategorySeriesDescription.$t = /*@__PURE__*/ markType(CategorySeriesDescription, 'CategorySeriesDescription', MarkerSeriesDescription.$);
    return CategorySeriesDescription;
})();
