// igniteui-react-core/esm2015/lib/CategoryChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { XYChartDescription } from "./XYChartDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CategoryChartDescription = /*@__PURE__*/ (() => {
    class CategoryChartDescription extends XYChartDescription {
        get tooltipTemplateRef() {
            return this.x6;
        }
        set tooltipTemplateRef(a) {
            this.x6 = a;
            this.g("TooltipTemplateRef");
        }
        get width() {
            return this.yb;
        }
        set width(a) {
            this.yb = a;
            this.g("Width");
        }
        get height() {
            return this.x4;
        }
        set height(a) {
            this.x4 = a;
            this.g("Height");
        }
        get background() {
            return this.x2;
        }
        set background(a) {
            this.x2 = a;
            this.g("Background");
        }
        get_type() {
            return "CategoryChart";
        }
        constructor() {
            super();
            this.x6 = null;
            this.yb = null;
            this.x4 = null;
            this.x2 = null;
            this.xv = 0;
            this.xw = 0;
            this.x7 = null;
            this.ya = null;
            this.x3 = null;
            this.x5 = null;
            this.wd = false;
            this.we = false;
            this.x8 = null;
            this.x9 = null;
            this.w0 = 0;
            this.w3 = 0;
            this.wz = 0;
            this.w2 = 0;
            this.w1 = 0;
            this.w4 = 0;
            this.xa = 0;
            this.x1 = null;
            this.wf = false;
            this.wg = false;
            this.wx = 0;
            this.wy = 0;
            this.wk = false;
            this.wh = false;
            this.wj = false;
            this.yc = null;
            this.wl = false;
            this.xx = 0;
            this.xc = 0;
            this.xb = 0;
            this.xd = 0;
            this.v7 = null;
            this.v8 = null;
            this.wi = false;
            this.w5 = 0;
            this.w6 = 0;
            this.w7 = 0;
            this.w8 = 0;
            this.w9 = 0;
            this.wb = false;
            this.wc = false;
        }
        get transitionInDuration() {
            return this.xv;
        }
        set transitionInDuration(a) {
            this.xv = a;
            this.g("TransitionInDuration");
        }
        get transitionOutDuration() {
            return this.xw;
        }
        set transitionOutDuration(a) {
            this.xw = a;
            this.g("TransitionOutDuration");
        }
        get transitionInEasingFunctionRef() {
            return this.x7;
        }
        set transitionInEasingFunctionRef(a) {
            this.x7 = a;
            this.g("TransitionInEasingFunctionRef");
        }
        get transitionOutEasingFunctionRef() {
            return this.ya;
        }
        set transitionOutEasingFunctionRef(a) {
            this.ya = a;
            this.g("TransitionOutEasingFunctionRef");
        }
        get chartType() {
            return this.x3;
        }
        set chartType(a) {
            this.x3 = a;
            this.g("ChartType");
        }
        get markerCollisionAvoidance() {
            return this.x5;
        }
        set markerCollisionAvoidance(a) {
            this.x5 = a;
            this.g("MarkerCollisionAvoidance");
        }
        get isSplineShapePartOfRange() {
            return this.wd;
        }
        set isSplineShapePartOfRange(a) {
            this.wd = a;
            this.g("IsSplineShapePartOfRange");
        }
        get isTransitionInEnabled() {
            return this.we;
        }
        set isTransitionInEnabled(a) {
            this.we = a;
            this.g("IsTransitionInEnabled");
        }
        get transitionInMode() {
            return this.x8;
        }
        set transitionInMode(a) {
            this.x8 = a;
            this.g("TransitionInMode");
        }
        get transitionInSpeedType() {
            return this.x9;
        }
        set transitionInSpeedType(a) {
            this.x9 = a;
            this.g("TransitionInSpeedType");
        }
        get xAxisInterval() {
            return this.w0;
        }
        set xAxisInterval(a) {
            this.w0 = a;
            this.g("XAxisInterval");
        }
        get xAxisMinorInterval() {
            return this.w3;
        }
        set xAxisMinorInterval(a) {
            this.w3 = a;
            this.g("XAxisMinorInterval");
        }
        get xAxisGap() {
            return this.wz;
        }
        set xAxisGap(a) {
            this.wz = a;
            this.g("XAxisGap");
        }
        get xAxisMinimumGapSize() {
            return this.w2;
        }
        set xAxisMinimumGapSize(a) {
            this.w2 = a;
            this.g("XAxisMinimumGapSize");
        }
        get xAxisMaximumGap() {
            return this.w1;
        }
        set xAxisMaximumGap(a) {
            this.w1 = a;
            this.g("XAxisMaximumGap");
        }
        get xAxisOverlap() {
            return this.w4;
        }
        set xAxisOverlap(a) {
            this.w4 = a;
            this.g("XAxisOverlap");
        }
        get yAxisInterval() {
            return this.xa;
        }
        set yAxisInterval(a) {
            this.xa = a;
            this.g("YAxisInterval");
        }
        get autoMarginAndAngleUpdateMode() {
            return this.x1;
        }
        set autoMarginAndAngleUpdateMode(a) {
            this.x1 = a;
            this.g("AutoMarginAndAngleUpdateMode");
        }
        get shouldAutoExpandMarginForInitialLabels() {
            return this.wf;
        }
        set shouldAutoExpandMarginForInitialLabels(a) {
            this.wf = a;
            this.g("ShouldAutoExpandMarginForInitialLabels");
        }
        get shouldConsiderAutoRotationForInitialLabels() {
            return this.wg;
        }
        set shouldConsiderAutoRotationForInitialLabels(a) {
            this.wg = a;
            this.g("ShouldConsiderAutoRotationForInitialLabels");
        }
        get autoExpandMarginExtraPadding() {
            return this.wx;
        }
        set autoExpandMarginExtraPadding(a) {
            this.wx = a;
            this.g("AutoExpandMarginExtraPadding");
        }
        get autoExpandMarginMaximumValue() {
            return this.wy;
        }
        set autoExpandMarginMaximumValue(a) {
            this.wy = a;
            this.g("AutoExpandMarginMaximumValue");
        }
        get yAxisFavorLabellingScaleEnd() {
            return this.wk;
        }
        set yAxisFavorLabellingScaleEnd(a) {
            this.wk = a;
            this.g("YAxisFavorLabellingScaleEnd");
        }
        get xAxisEnhancedIntervalPreferMoreCategoryLabels() {
            return this.wh;
        }
        set xAxisEnhancedIntervalPreferMoreCategoryLabels(a) {
            this.wh = a;
            this.g("XAxisEnhancedIntervalPreferMoreCategoryLabels");
        }
        get yAxisEnhancedIntervalPreferMoreCategoryLabels() {
            return this.wj;
        }
        set yAxisEnhancedIntervalPreferMoreCategoryLabels(a) {
            this.wj = a;
            this.g("YAxisEnhancedIntervalPreferMoreCategoryLabels");
        }
        get yAxisAutoRangeBufferMode() {
            return this.yc;
        }
        set yAxisAutoRangeBufferMode(a) {
            this.yc = a;
            this.g("YAxisAutoRangeBufferMode");
        }
        get yAxisIsLogarithmic() {
            return this.wl;
        }
        set yAxisIsLogarithmic(a) {
            this.wl = a;
            this.g("YAxisIsLogarithmic");
        }
        get yAxisLogarithmBase() {
            return this.xx;
        }
        set yAxisLogarithmBase(a) {
            this.xx = a;
            this.g("YAxisLogarithmBase");
        }
        get yAxisMinimumValue() {
            return this.xc;
        }
        set yAxisMinimumValue(a) {
            this.xc = a;
            this.g("YAxisMinimumValue");
        }
        get yAxisMaximumValue() {
            return this.xb;
        }
        set yAxisMaximumValue(a) {
            this.xb = a;
            this.g("YAxisMaximumValue");
        }
        get yAxisMinorInterval() {
            return this.xd;
        }
        set yAxisMinorInterval(a) {
            this.xd = a;
            this.g("YAxisMinorInterval");
        }
        get negativeBrushes() {
            return this.v7;
        }
        set negativeBrushes(a) {
            this.v7 = a;
            this.g("NegativeBrushes");
        }
        get negativeOutlines() {
            return this.v8;
        }
        set negativeOutlines(a) {
            this.v8 = a;
            this.g("NegativeOutlines");
        }
        get yAxisAbbreviateLargeNumbers() {
            return this.wi;
        }
        set yAxisAbbreviateLargeNumbers(a) {
            this.wi = a;
            this.g("YAxisAbbreviateLargeNumbers");
        }
        get xAxisZoomMaximumCategoryRange() {
            return this.w5;
        }
        set xAxisZoomMaximumCategoryRange(a) {
            this.w5 = a;
            this.g("XAxisZoomMaximumCategoryRange");
        }
        get xAxisZoomMaximumItemSpan() {
            return this.w6;
        }
        set xAxisZoomMaximumItemSpan(a) {
            this.w6 = a;
            this.g("XAxisZoomMaximumItemSpan");
        }
        get xAxisZoomToCategoryRange() {
            return this.w7;
        }
        set xAxisZoomToCategoryRange(a) {
            this.w7 = a;
            this.g("XAxisZoomToCategoryRange");
        }
        get xAxisZoomToCategoryStart() {
            return this.w8;
        }
        set xAxisZoomToCategoryStart(a) {
            this.w8 = a;
            this.g("XAxisZoomToCategoryStart");
        }
        get xAxisZoomToItemSpan() {
            return this.w9;
        }
        set xAxisZoomToItemSpan(a) {
            this.w9 = a;
            this.g("XAxisZoomToItemSpan");
        }
        get isCategoryHighlightingEnabled() {
            return this.wb;
        }
        set isCategoryHighlightingEnabled(a) {
            this.wb = a;
            this.g("IsCategoryHighlightingEnabled");
        }
        get isItemHighlightingEnabled() {
            return this.wc;
        }
        set isItemHighlightingEnabled(a) {
            this.wc = a;
            this.g("IsItemHighlightingEnabled");
        }
    }
    CategoryChartDescription.$t = /*@__PURE__*/ markType(CategoryChartDescription, 'CategoryChartDescription', XYChartDescription.$);
    return CategoryChartDescription;
})();
