// igniteui-react-core/esm2015/lib/CategoryAngleAxisDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategoryAxisBaseDescription } from "./CategoryAxisBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CategoryAngleAxisDescription = /*@__PURE__*/ (() => {
    class CategoryAngleAxisDescription extends CategoryAxisBaseDescription {
        get_type() {
            return "CategoryAngleAxis";
        }
        constructor() {
            super();
            this.d7 = 0;
            this.ee = null;
            this.d5 = 0;
            this.d3 = 0;
            this.d6 = 0;
            this.d4 = 0;
            this.ed = null;
        }
        get startAngleOffset() {
            return this.d7;
        }
        set startAngleOffset(a) {
            this.d7 = a;
            this.g("StartAngleOffset");
        }
        get labelMode() {
            return this.ee;
        }
        set labelMode(a) {
            this.ee = a;
            this.g("LabelMode");
        }
        get interval() {
            return this.d5;
        }
        set interval(a) {
            this.d5 = a;
            this.g("Interval");
        }
        get actualInterval() {
            return this.d3;
        }
        set actualInterval(a) {
            this.d3 = a;
            this.g("ActualInterval");
        }
        get minorInterval() {
            return this.d6;
        }
        set minorInterval(a) {
            this.d6 = a;
            this.g("MinorInterval");
        }
        get actualMinorInterval() {
            return this.d4;
        }
        set actualMinorInterval(a) {
            this.d4 = a;
            this.g("ActualMinorInterval");
        }
        get actualIntervalChangeRef() {
            return this.ed;
        }
        set actualIntervalChangeRef(a) {
            this.ed = a;
            this.g("ActualIntervalChangeRef");
        }
    }
    CategoryAngleAxisDescription.$t = /*@__PURE__*/ markType(CategoryAngleAxisDescription, 'CategoryAngleAxisDescription', CategoryAxisBaseDescription.$);
    return CategoryAngleAxisDescription;
})();
