// igniteui-react-core/esm2015/lib/CategoryItemHighlightLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CategoryItemHighlightLayerDescription = /*@__PURE__*/ (() => {
    class CategoryItemHighlightLayerDescription extends AnnotationLayerDescription {
        get_type() {
            return "CategoryItemHighlightLayer";
        }
        constructor() {
            super();
            this.gl = null;
            this.gk = null;
            this.ga = false;
            this.gf = null;
            this.gj = null;
            this.gg = null;
            this.gh = null;
            this.gi = null;
            this.gd = 0;
            this.f9 = false;
        }
        get targetSeriesRef() {
            return this.gl;
        }
        set targetSeriesRef(a) {
            this.gl = a;
            this.g("TargetSeriesRef");
        }
        get targetSeriesName() {
            return this.gk;
        }
        set targetSeriesName(a) {
            this.gk = a;
            this.g("TargetSeriesName");
        }
        get useInterpolation() {
            return this.ga;
        }
        set useInterpolation(a) {
            this.ga = a;
            this.g("UseInterpolation");
        }
        get highlightType() {
            return this.gf;
        }
        set highlightType(a) {
            this.gf = a;
            this.g("HighlightType");
        }
        get markerType() {
            return this.gj;
        }
        set markerType(a) {
            this.gj = a;
            this.g("MarkerType");
        }
        get markerBrush() {
            return this.gg;
        }
        set markerBrush(a) {
            this.gg = a;
            this.g("MarkerBrush");
        }
        get markerOutline() {
            return this.gh;
        }
        set markerOutline(a) {
            this.gh = a;
            this.g("MarkerOutline");
        }
        get markerTemplateRef() {
            return this.gi;
        }
        set markerTemplateRef(a) {
            this.gi = a;
            this.g("MarkerTemplateRef");
        }
        get bandHighlightWidth() {
            return this.gd;
        }
        set bandHighlightWidth(a) {
            this.gd = a;
            this.g("BandHighlightWidth");
        }
        get skipUnknownValues() {
            return this.f9;
        }
        set skipUnknownValues(a) {
            this.f9 = a;
            this.g("SkipUnknownValues");
        }
    }
    CategoryItemHighlightLayerDescription.$t = /*@__PURE__*/ markType(CategoryItemHighlightLayerDescription, 'CategoryItemHighlightLayerDescription', AnnotationLayerDescription.$);
    return CategoryItemHighlightLayerDescription;
})();
