// igniteui-react-core/esm2015/lib/CellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CellInfoDescription = /*@__PURE__*/ (() => {
    class CellInfoDescription extends Description {
        get_type() {
            return "CellInfo";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.eh = null;
            this.b8 = 0;
            this.y_1 = false;
            this.ac = false;
            this.m = false;
            this.k = false;
            this.w = false;
            this.ad = false;
            this.n = false;
            this.x_1 = false;
            this.h = null;
            this.s = false;
            this.z = false;
            this.aa = false;
            this.v = false;
            this.r = false;
            this.cf = 0;
            this.ec = null;
            this.bb = 0;
            this.bc = 0;
            this.ck = 0;
            this.cl = 0;
            this.a4 = 0;
            this.a2 = 0;
            this.az = 0;
            this.a7 = 0;
            this.a1 = 0;
            this.ce = 0;
            this.cm = 0;
            this.ab = false;
            this.a9 = 0;
            this.ay = 0;
            this.a6 = 0;
            this.t = false;
            this.d3 = null;
            this.ej = null;
            this.dr = null;
            this.ds = null;
            this.dk = null;
            this.dl = null;
            this.dj = null;
            this.dy = null;
            this.ed = null;
            this.d4 = null;
            this.d5 = null;
            this.ef = null;
            this.eg = null;
            this.ea = null;
            this.d6 = null;
            this.dz = null;
            this.a8 = 0;
            this.d8 = null;
            this.eb = null;
            this.ei = null;
            this.dq = null;
            this.dt = null;
            this.d0 = null;
            this.a5 = 0;
            this.d1 = null;
            this.d2 = null;
            this.dm = null;
            this.a0 = 0;
            this.dn = null;
            this.dp = null;
            this.d7 = null;
            this.ba = 0;
            this.ch = 0;
            this.cj = 0;
            this.ci = 0;
            this.cg = 0;
            this.b1 = 0;
            this.b3 = 0;
            this.b2 = 0;
            this.b0 = 0;
            this.b5 = 0;
            this.b7 = 0;
            this.b6 = 0;
            this.b4 = 0;
            this.bt = 0;
            this.bv = 0;
            this.bu = 0;
            this.bs = 0;
            this.cb = 0;
            this.cd = 0;
            this.cc = 0;
            this.ca = 0;
            this.bx = 0;
            this.bz = 0;
            this.by = 0;
            this.bw = 0;
            this.ee = null;
            this.q = false;
            this.l = false;
            this.d9 = null;
            this.dv = null;
            this.a3 = 0;
            this.dw = null;
            this.dx = null;
            this.p = false;
            this.o = false;
            this.du = null;
            this.u = false;
            this.b9 = 0;
        }
        get styleKey() {
            return this.eh;
        }
        set styleKey(a) {
            this.eh = a;
            this.g("StyleKey");
        }
        get dataRow() {
            return this.b8;
        }
        set dataRow(a) {
            this.b8 = a;
            this.g("DataRow");
        }
        get isPositionDirty() {
            return this.y_1;
        }
        set isPositionDirty(a) {
            this.y_1 = a;
            this.g("IsPositionDirty");
        }
        get isSizeDirty() {
            return this.ac;
        }
        set isSizeDirty(a) {
            this.ac = a;
            this.g("IsSizeDirty");
        }
        get isContentDirty() {
            return this.m;
        }
        set isContentDirty(a) {
            this.m = a;
            this.g("IsContentDirty");
        }
        get isBorderDirty() {
            return this.k;
        }
        set isBorderDirty(a) {
            this.k = a;
            this.g("IsBorderDirty");
        }
        get isLayerDirty() {
            return this.w;
        }
        set isLayerDirty(a) {
            this.w = a;
            this.g("IsLayerDirty");
        }
        get isStateDirty() {
            return this.ad;
        }
        set isStateDirty(a) {
            this.ad = a;
            this.g("IsStateDirty");
        }
        get isDataDirty() {
            return this.n;
        }
        set isDataDirty(a) {
            this.n = a;
            this.g("IsDataDirty");
        }
        get isPlaceholdContentNeeded() {
            return this.x_1;
        }
        set isPlaceholdContentNeeded(a) {
            this.x_1 = a;
            this.g("IsPlaceholdContentNeeded");
        }
        get actionManager() {
            return this.h;
        }
        set actionManager(a) {
            this.h = a;
            this.g("ActionManager");
        }
        get isHitTestVisible() {
            return this.s;
        }
        set isHitTestVisible(a) {
            this.s = a;
            this.g("IsHitTestVisible");
        }
        get isRowPinned() {
            return this.z;
        }
        set isRowPinned(a) {
            this.z = a;
            this.g("IsRowPinned");
        }
        get isRowSticky() {
            return this.aa;
        }
        set isRowSticky(a) {
            this.aa = a;
            this.g("IsRowSticky");
        }
        get isLastStickyRow() {
            return this.v;
        }
        set isLastStickyRow(a) {
            this.v = a;
            this.g("IsLastStickyRow");
        }
        get isFilterRow() {
            return this.r;
        }
        set isFilterRow(a) {
            this.r = a;
            this.g("IsFilterRow");
        }
        get indent() {
            return this.cf;
        }
        set indent(a) {
            this.cf = a;
            this.g("Indent");
        }
        get rowItemRef() {
            return this.ec;
        }
        set rowItemRef(a) {
            this.ec = a;
            this.g("RowItemRef");
        }
        get x() {
            return this.bb;
        }
        set x(a) {
            this.bb = a;
            this.g("X");
        }
        get y() {
            return this.bc;
        }
        set y(a) {
            this.bc = a;
            this.g("Y");
        }
        get snappedX() {
            return this.ck;
        }
        set snappedX(a) {
            this.ck = a;
            this.g("SnappedX");
        }
        get snappedY() {
            return this.cl;
        }
        set snappedY(a) {
            this.cl = a;
            this.g("SnappedY");
        }
        get editOpacity() {
            return this.a4;
        }
        set editOpacity(a) {
            this.a4 = a;
            this.g("EditOpacity");
        }
        get contentOpacity() {
            return this.a2;
        }
        set contentOpacity(a) {
            this.a2 = a;
            this.g("ContentOpacity");
        }
        get actualContentOpacity() {
            return this.az;
        }
        set actualContentOpacity(a) {
            this.az = a;
            this.g("ActualContentOpacity");
        }
        get opacity() {
            return this.a7;
        }
        set opacity(a) {
            this.a7 = a;
            this.g("Opacity");
        }
        get actualOpacity() {
            return this.a1;
        }
        set actualOpacity(a) {
            this.a1 = a;
            this.g("ActualOpacity");
        }
        get height() {
            return this.ce;
        }
        set height(a) {
            this.ce = a;
            this.g("Height");
        }
        get width() {
            return this.cm;
        }
        set width(a) {
            this.cm = a;
            this.g("Width");
        }
        get isSelected() {
            return this.ab;
        }
        set isSelected(a) {
            this.ab = a;
            this.g("IsSelected");
        }
        get selectedStatus() {
            return this.a9;
        }
        set selectedStatus(a) {
            this.a9 = a;
            this.g("SelectedStatus");
        }
        get activationStatus() {
            return this.ay;
        }
        set activationStatus(a) {
            this.ay = a;
            this.g("ActivationStatus");
        }
        get hoverStatus() {
            return this.a6;
        }
        set hoverStatus(a) {
            this.a6 = a;
            this.g("HoverStatus");
        }
        get isHoverable() {
            return this.t;
        }
        set isHoverable(a) {
            this.t = a;
            this.g("IsHoverable");
        }
        get horizontalAlignment() {
            return this.d3;
        }
        set horizontalAlignment(a) {
            this.d3 = a;
            this.g("HorizontalAlignment");
        }
        get verticalAlignment() {
            return this.ej;
        }
        set verticalAlignment(a) {
            this.ej = a;
            this.g("VerticalAlignment");
        }
        get background() {
            return this.dr;
        }
        set background(a) {
            this.dr = a;
            this.g("Background");
        }
        get border() {
            return this.ds;
        }
        set border(a) {
            this.ds = a;
            this.g("Border");
        }
        get actualBackground() {
            return this.dk;
        }
        set actualBackground(a) {
            this.dk = a;
            this.g("ActualBackground");
        }
        get actualBorder() {
            return this.dl;
        }
        set actualBorder(a) {
            this.dl = a;
            this.g("ActualBorder");
        }
        get activationBorder() {
            return this.dj;
        }
        set activationBorder(a) {
            this.dj = a;
            this.g("ActivationBorder");
        }
        get errorBorder() {
            return this.dy;
        }
        set errorBorder(a) {
            this.dy = a;
            this.g("ErrorBorder");
        }
        get selectedBackground() {
            return this.ed;
        }
        set selectedBackground(a) {
            this.ed = a;
            this.g("SelectedBackground");
        }
        get hoverBackground() {
            return this.d4;
        }
        set hoverBackground(a) {
            this.d4 = a;
            this.g("HoverBackground");
        }
        get hoverTextColor() {
            return this.d5;
        }
        set hoverTextColor(a) {
            this.d5 = a;
            this.g("HoverTextColor");
        }
        get sortIndicatorColor() {
            return this.ef;
        }
        set sortIndicatorColor(a) {
            this.ef = a;
            this.g("SortIndicatorColor");
        }
        get stickyRowBackground() {
            return this.eg;
        }
        set stickyRowBackground(a) {
            this.eg = a;
            this.g("StickyRowBackground");
        }
        get pinnedRowBackground() {
            return this.ea;
        }
        set pinnedRowBackground(a) {
            this.ea = a;
            this.g("PinnedRowBackground");
        }
        get lastStickyRowBackground() {
            return this.d6;
        }
        set lastStickyRowBackground(a) {
            this.d6 = a;
            this.g("LastStickyRowBackground");
        }
        get filterRowBackground() {
            return this.dz;
        }
        set filterRowBackground(a) {
            this.dz = a;
            this.g("FilterRowBackground");
        }
        get pinnedRowOpacity() {
            return this.a8;
        }
        set pinnedRowOpacity(a) {
            this.a8 = a;
            this.g("PinnedRowOpacity");
        }
        get originalValueRef() {
            return this.d8;
        }
        set originalValueRef(a) {
            this.d8 = a;
            this.g("OriginalValueRef");
        }
        get renderValue() {
            return this.eb;
        }
        set renderValue(a) {
            this.eb = a;
            this.g("RenderValue");
        }
        get textColor() {
            return this.ei;
        }
        set textColor(a) {
            this.ei = a;
            this.g("TextColor");
        }
        get actualTextColor() {
            return this.dq;
        }
        set actualTextColor(a) {
            this.dq = a;
            this.g("ActualTextColor");
        }
        get deletedTextColor() {
            return this.dt;
        }
        set deletedTextColor(a) {
            this.dt = a;
            this.g("DeletedTextColor");
        }
        get fontFamily() {
            return this.d0;
        }
        set fontFamily(a) {
            this.d0 = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.a5;
        }
        set fontSize(a) {
            this.a5 = a;
            this.g("FontSize");
        }
        get fontStyle() {
            return this.d1;
        }
        set fontStyle(a) {
            this.d1 = a;
            this.g("FontStyle");
        }
        get fontWeight() {
            return this.d2;
        }
        set fontWeight(a) {
            this.d2 = a;
            this.g("FontWeight");
        }
        get actualFontFamily() {
            return this.dm;
        }
        set actualFontFamily(a) {
            this.dm = a;
            this.g("ActualFontFamily");
        }
        get actualFontSize() {
            return this.a0;
        }
        set actualFontSize(a) {
            this.a0 = a;
            this.g("ActualFontSize");
        }
        get actualFontStyle() {
            return this.dn;
        }
        set actualFontStyle(a) {
            this.dn = a;
            this.g("ActualFontStyle");
        }
        get actualFontWeight() {
            return this.dp;
        }
        set actualFontWeight(a) {
            this.dp = a;
            this.g("ActualFontWeight");
        }
        get lineBreakMode() {
            return this.d7;
        }
        set lineBreakMode(a) {
            this.d7 = a;
            this.g("LineBreakMode");
        }
        get virtualizationPercentage() {
            return this.ba;
        }
        set virtualizationPercentage(a) {
            this.ba = a;
            this.g("VirtualizationPercentage");
        }
        get paddingLeft() {
            return this.ch;
        }
        set paddingLeft(a) {
            this.ch = a;
            this.g("PaddingLeft");
        }
        get paddingTop() {
            return this.cj;
        }
        set paddingTop(a) {
            this.cj = a;
            this.g("PaddingTop");
        }
        get paddingRight() {
            return this.ci;
        }
        set paddingRight(a) {
            this.ci = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.cg;
        }
        set paddingBottom(a) {
            this.cg = a;
            this.g("PaddingBottom");
        }
        get actualPaddingLeft() {
            return this.b1;
        }
        set actualPaddingLeft(a) {
            this.b1 = a;
            this.g("ActualPaddingLeft");
        }
        get actualPaddingTop() {
            return this.b3;
        }
        set actualPaddingTop(a) {
            this.b3 = a;
            this.g("ActualPaddingTop");
        }
        get actualPaddingRight() {
            return this.b2;
        }
        set actualPaddingRight(a) {
            this.b2 = a;
            this.g("ActualPaddingRight");
        }
        get actualPaddingBottom() {
            return this.b0;
        }
        set actualPaddingBottom(a) {
            this.b0 = a;
            this.g("ActualPaddingBottom");
        }
        get borderLeftWidth() {
            return this.b5;
        }
        set borderLeftWidth(a) {
            this.b5 = a;
            this.g("BorderLeftWidth");
        }
        get borderTopWidth() {
            return this.b7;
        }
        set borderTopWidth(a) {
            this.b7 = a;
            this.g("BorderTopWidth");
        }
        get borderRightWidth() {
            return this.b6;
        }
        set borderRightWidth(a) {
            this.b6 = a;
            this.g("BorderRightWidth");
        }
        get borderBottomWidth() {
            return this.b4;
        }
        set borderBottomWidth(a) {
            this.b4 = a;
            this.g("BorderBottomWidth");
        }
        get activationBorderLeftWidth() {
            return this.bt;
        }
        set activationBorderLeftWidth(a) {
            this.bt = a;
            this.g("ActivationBorderLeftWidth");
        }
        get activationBorderTopWidth() {
            return this.bv;
        }
        set activationBorderTopWidth(a) {
            this.bv = a;
            this.g("ActivationBorderTopWidth");
        }
        get activationBorderRightWidth() {
            return this.bu;
        }
        set activationBorderRightWidth(a) {
            this.bu = a;
            this.g("ActivationBorderRightWidth");
        }
        get activationBorderBottomWidth() {
            return this.bs;
        }
        set activationBorderBottomWidth(a) {
            this.bs = a;
            this.g("ActivationBorderBottomWidth");
        }
        get errorBorderLeftWidth() {
            return this.cb;
        }
        set errorBorderLeftWidth(a) {
            this.cb = a;
            this.g("ErrorBorderLeftWidth");
        }
        get errorBorderTopWidth() {
            return this.cd;
        }
        set errorBorderTopWidth(a) {
            this.cd = a;
            this.g("ErrorBorderTopWidth");
        }
        get errorBorderRightWidth() {
            return this.cc;
        }
        set errorBorderRightWidth(a) {
            this.cc = a;
            this.g("ErrorBorderRightWidth");
        }
        get errorBorderBottomWidth() {
            return this.ca;
        }
        set errorBorderBottomWidth(a) {
            this.ca = a;
            this.g("ErrorBorderBottomWidth");
        }
        get actualBorderLeftWidth() {
            return this.bx;
        }
        set actualBorderLeftWidth(a) {
            this.bx = a;
            this.g("ActualBorderLeftWidth");
        }
        get actualBorderTopWidth() {
            return this.bz;
        }
        set actualBorderTopWidth(a) {
            this.bz = a;
            this.g("ActualBorderTopWidth");
        }
        get actualBorderRightWidth() {
            return this.by;
        }
        set actualBorderRightWidth(a) {
            this.by = a;
            this.g("ActualBorderRightWidth");
        }
        get actualBorderBottomWidth() {
            return this.bw;
        }
        set actualBorderBottomWidth(a) {
            this.bw = a;
            this.g("ActualBorderBottomWidth");
        }
        get sortDirection() {
            return this.ee;
        }
        set sortDirection(a) {
            this.ee = a;
            this.g("SortDirection");
        }
        get isExpanded() {
            return this.q;
        }
        set isExpanded(a) {
            this.q = a;
            this.g("IsExpanded");
        }
        get isCollapsable() {
            return this.l;
        }
        set isCollapsable(a) {
            this.l = a;
            this.g("IsCollapsable");
        }
        get pinned() {
            return this.d9;
        }
        set pinned(a) {
            this.d9 = a;
            this.g("Pinned");
        }
        get editFontFamily() {
            return this.dv;
        }
        set editFontFamily(a) {
            this.dv = a;
            this.g("EditFontFamily");
        }
        get editFontSize() {
            return this.a3;
        }
        set editFontSize(a) {
            this.a3 = a;
            this.g("EditFontSize");
        }
        get editFontStyle() {
            return this.dw;
        }
        set editFontStyle(a) {
            this.dw = a;
            this.g("EditFontStyle");
        }
        get editFontWeight() {
            return this.dx;
        }
        set editFontWeight(a) {
            this.dx = a;
            this.g("EditFontWeight");
        }
        get isEdited() {
            return this.p;
        }
        set isEdited(a) {
            this.p = a;
            this.g("IsEdited");
        }
        get isDeleted() {
            return this.o;
        }
        set isDeleted(a) {
            this.o = a;
            this.g("IsDeleted");
        }
        get editError() {
            return this.du;
        }
        set editError(a) {
            this.du = a;
            this.g("EditError");
        }
        get isInEditMode() {
            return this.u;
        }
        set isInEditMode(a) {
            this.u = a;
            this.g("IsInEditMode");
        }
        get editID() {
            return this.b9;
        }
        set editID(a) {
            this.b9 = a;
            this.g("EditID");
        }
    }
    CellInfoDescription.$t = /*@__PURE__*/ markType(CellInfoDescription, 'CellInfoDescription', Description.$);
    CellInfoDescription.__marshalByValue = true;
    CellInfoDescription.__marshalByValueAlias = "CellModel";
    return CellInfoDescription;
})();
