// igniteui-react-core/esm2015/lib/CellActionManagerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CellActionManagerDescription = /*@__PURE__*/ (() => {
    class CellActionManagerDescription extends Description {
        get_type() {
            return "CellActionManager";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = false;
            this.i = false;
        }
        get isControlPressed() {
            return this.h;
        }
        set isControlPressed(a) {
            this.h = a;
            this.g("IsControlPressed");
        }
        get isShiftPressed() {
            return this.i;
        }
        set isShiftPressed(a) {
            this.i = a;
            this.g("IsShiftPressed");
        }
    }
    CellActionManagerDescription.$t = /*@__PURE__*/ markType(CellActionManagerDescription, 'CellActionManagerDescription', Description.$);
    return CellActionManagerDescription;
})();
