// igniteui-react-core/esm2015/lib/CellStyleRequestedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CellStyleRequestedEventArgsDescription = /*@__PURE__*/ (() => {
    class CellStyleRequestedEventArgsDescription extends Description {
        get_type() {
            return "CellStyleRequestedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = null;
            this.j = null;
            this.h = 0;
        }
        get styleKey() {
            return this.k;
        }
        set styleKey(a) {
            this.k = a;
            this.g("StyleKey");
        }
        get resolvedValueRef() {
            return this.j;
        }
        set resolvedValueRef(a) {
            this.j = a;
            this.g("ResolvedValueRef");
        }
        get rowNumber() {
            return this.h;
        }
        set rowNumber(a) {
            this.h = a;
            this.g("RowNumber");
        }
    }
    CellStyleRequestedEventArgsDescription.$t = /*@__PURE__*/ markType(CellStyleRequestedEventArgsDescription, 'CellStyleRequestedEventArgsDescription', Description.$);
    return CellStyleRequestedEventArgsDescription;
})();
