// igniteui-react-core/esm2015/lib/ChartSummaryDescriptionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ChartSummaryDescriptionDescription = /*@__PURE__*/ (() => {
    class ChartSummaryDescriptionDescription extends Description {
        get_type() {
            return "ChartSummaryDescription";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.m = null;
            this.j = null;
            this.k = null;
            this.n = null;
        }
        get field() {
            return this.l;
        }
        set field(a) {
            this.l = a;
            this.g("Field");
        }
        get operand() {
            return this.m;
        }
        set operand(a) {
            this.m = a;
            this.g("Operand");
        }
        get alias() {
            return this.j;
        }
        set alias(a) {
            this.j = a;
            this.g("Alias");
        }
        get calculatorDisplayName() {
            return this.k;
        }
        set calculatorDisplayName(a) {
            this.k = a;
            this.g("CalculatorDisplayName");
        }
        get provideCalculatorRef() {
            return this.n;
        }
        set provideCalculatorRef(a) {
            this.n = a;
            this.g("ProvideCalculatorRef");
        }
    }
    ChartSummaryDescriptionDescription.$t = /*@__PURE__*/ markType(ChartSummaryDescriptionDescription, 'ChartSummaryDescriptionDescription', Description.$);
    ChartSummaryDescriptionDescription.__marshalByValue = true;
    ChartSummaryDescriptionDescription.__marshalByValueAlias = "ChartSummaryDescription";
    return ChartSummaryDescriptionDescription;
})();
