// igniteui-react-core/esm2015/lib/ChartCursorEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ChartCursorEventArgsDescription = /*@__PURE__*/ (() => {
    class ChartCursorEventArgsDescription extends Description {
        get_type() {
            return "ChartCursorEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.h = null;
            this.j = null;
        }
        get itemRef() {
            return this.l;
        }
        set itemRef(a) {
            this.l = a;
            this.g("ItemRef");
        }
        get series() {
            return this.h;
        }
        set series(a) {
            this.h = a;
            this.g("Series");
        }
        get seriesViewer() {
            return this.j;
        }
        set seriesViewer(a) {
            this.j = a;
            this.g("SeriesViewer");
        }
    }
    ChartCursorEventArgsDescription.$t = /*@__PURE__*/ markType(ChartCursorEventArgsDescription, 'ChartCursorEventArgsDescription', Description.$);
    return ChartCursorEventArgsDescription;
})();
