// igniteui-react-core/esm2015/lib/CheckedChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CheckedChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class CheckedChangedEventArgsDescription extends Description {
        get_type() {
            return "CheckedChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = 0;
            this.h = false;
            this.l = null;
        }
        get index() {
            return this.j;
        }
        set index(a) {
            this.j = a;
            this.g("Index");
        }
        get isChecked() {
            return this.h;
        }
        set isChecked(a) {
            this.h = a;
            this.g("IsChecked");
        }
        get primaryKeyRef() {
            return this.l;
        }
        set primaryKeyRef(a) {
            this.l = a;
            this.g("PrimaryKeyRef");
        }
    }
    CheckedChangedEventArgsDescription.$t = /*@__PURE__*/ markType(CheckedChangedEventArgsDescription, 'CheckedChangedEventArgsDescription', Description.$);
    return CheckedChangedEventArgsDescription;
})();
