// igniteui-react-core/esm2015/lib/CheckboxChangeEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CheckboxChangeEventArgsDescription = /*@__PURE__*/ (() => {
    class CheckboxChangeEventArgsDescription extends Description {
        get_type() {
            return "CheckboxChangeEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = false;
            this.i = false;
        }
        get isChecked() {
            return this.h;
        }
        set isChecked(a) {
            this.h = a;
            this.g("IsChecked");
        }
        get isIndeterminate() {
            return this.i;
        }
        set isIndeterminate(a) {
            this.i = a;
            this.g("IsIndeterminate");
        }
    }
    CheckboxChangeEventArgsDescription.$t = /*@__PURE__*/ markType(CheckboxChangeEventArgsDescription, 'CheckboxChangeEventArgsDescription', Description.$);
    return CheckboxChangeEventArgsDescription;
})();
