// igniteui-react-core/esm2015/lib/CheckboxDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CheckboxDescription = /*@__PURE__*/ (() => {
    class CheckboxDescription extends Description {
        get_type() {
            return "Checkbox";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.aq = null;
            this.a0 = null;
            this.am = null;
            this.as = null;
            this.az = null;
            this.al = null;
            this.a1 = null;
            this.at = null;
            this.an = null;
            this.ak = null;
            this.w = 0;
            this.t = 0;
            this.x = 0;
            this.y = 0;
            this.v = 0;
            this.u = 0;
            this.aj = null;
            this.av = null;
            this.au = null;
            this.aw = null;
            this.ay = null;
            this.ag = 0;
            this.af = 0;
            this.ap = null;
            this.ao = null;
            this.a2 = null;
            this.l = false;
            this.ax = null;
            this.j = false;
            this.h = false;
            this.m = false;
            this.i = false;
            this.k = false;
            this.ar = null;
        }
        get baseTheme() {
            return this.aq;
        }
        set baseTheme(a) {
            this.aq = a;
            this.g("BaseTheme");
        }
        get uncheckedBackgroundColor() {
            return this.a0;
        }
        set uncheckedBackgroundColor(a) {
            this.a0 = a;
            this.g("UncheckedBackgroundColor");
        }
        get actualUncheckedBackgroundColor() {
            return this.am;
        }
        set actualUncheckedBackgroundColor(a) {
            this.am = a;
            this.g("ActualUncheckedBackgroundColor");
        }
        get checkedBackgroundColor() {
            return this.as;
        }
        set checkedBackgroundColor(a) {
            this.as = a;
            this.g("CheckedBackgroundColor");
        }
        get tickColor() {
            return this.az;
        }
        set tickColor(a) {
            this.az = a;
            this.g("TickColor");
        }
        get actualTickColor() {
            return this.al;
        }
        set actualTickColor(a) {
            this.al = a;
            this.g("ActualTickColor");
        }
        get uncheckedBorderColor() {
            return this.a1;
        }
        set uncheckedBorderColor(a) {
            this.a1 = a;
            this.g("UncheckedBorderColor");
        }
        get checkedBorderColor() {
            return this.at;
        }
        set checkedBorderColor(a) {
            this.at = a;
            this.g("CheckedBorderColor");
        }
        get actualUncheckedBorderColor() {
            return this.an;
        }
        set actualUncheckedBorderColor(a) {
            this.an = a;
            this.g("ActualUncheckedBorderColor");
        }
        get actualCheckedBorderColor() {
            return this.ak;
        }
        set actualCheckedBorderColor(a) {
            this.ak = a;
            this.g("ActualCheckedBorderColor");
        }
        get borderWidth() {
            return this.w;
        }
        set borderWidth(a) {
            this.w = a;
            this.g("BorderWidth");
        }
        get actualBorderWidth() {
            return this.t;
        }
        set actualBorderWidth(a) {
            this.t = a;
            this.g("ActualBorderWidth");
        }
        get cornerRadius() {
            return this.x;
        }
        set cornerRadius(a) {
            this.x = a;
            this.g("CornerRadius");
        }
        get tickStrokeWidth() {
            return this.y;
        }
        set tickStrokeWidth(a) {
            this.y = a;
            this.g("TickStrokeWidth");
        }
        get actualTickStrokeWidth() {
            return this.v;
        }
        set actualTickStrokeWidth(a) {
            this.v = a;
            this.g("ActualTickStrokeWidth");
        }
        get actualCornerRadius() {
            return this.u;
        }
        set actualCornerRadius(a) {
            this.u = a;
            this.g("ActualCornerRadius");
        }
        get actualCheckedBackgroundColor() {
            return this.aj;
        }
        set actualCheckedBackgroundColor(a) {
            this.aj = a;
            this.g("ActualCheckedBackgroundColor");
        }
        get inputId() {
            return this.av;
        }
        set inputId(a) {
            this.av = a;
            this.g("InputId");
        }
        get id() {
            return this.au;
        }
        set id(a) {
            this.au = a;
            this.g("Id");
        }
        get labelId() {
            return this.aw;
        }
        set labelId(a) {
            this.aw = a;
            this.g("LabelId");
        }
        get name() {
            return this.ay;
        }
        set name(a) {
            this.ay = a;
            this.g("Name");
        }
        get tabIndex() {
            return this.ag;
        }
        set tabIndex(a) {
            this.ag = a;
            this.g("TabIndex");
        }
        get required() {
            return this.af;
        }
        set required(a) {
            this.af = a;
            this.g("Required");
        }
        get ariaLabelledBy() {
            return this.ap;
        }
        set ariaLabelledBy(a) {
            this.ap = a;
            this.g("AriaLabelledBy");
        }
        get ariaLabel() {
            return this.ao;
        }
        set ariaLabel(a) {
            this.ao = a;
            this.g("AriaLabel");
        }
        get valueRef() {
            return this.a2;
        }
        set valueRef(a) {
            this.a2 = a;
            this.g("ValueRef");
        }
        get focused() {
            return this.l;
        }
        set focused(a) {
            this.l = a;
            this.g("Focused");
        }
        get labelPosition() {
            return this.ax;
        }
        set labelPosition(a) {
            this.ax = a;
            this.g("LabelPosition");
        }
        get disableRipple() {
            return this.j;
        }
        set disableRipple(a) {
            this.j = a;
            this.g("DisableRipple");
        }
        get checked() {
            return this.h;
        }
        set checked(a) {
            this.h = a;
            this.g("Checked");
        }
        get indeterminate() {
            return this.m;
        }
        set indeterminate(a) {
            this.m = a;
            this.g("Indeterminate");
        }
        get disabled() {
            return this.i;
        }
        set disabled(a) {
            this.i = a;
            this.g("Disabled");
        }
        get disableTransitions() {
            return this.k;
        }
        set disableTransitions(a) {
            this.k = a;
            this.g("DisableTransitions");
        }
        get changeRef() {
            return this.ar;
        }
        set changeRef(a) {
            this.ar = a;
            this.g("ChangeRef");
        }
    }
    CheckboxDescription.$t = /*@__PURE__*/ markType(CheckboxDescription, 'CheckboxDescription', Description.$);
    return CheckboxDescription;
})();
