// igniteui-react-core/esm2015/lib/CodeGenerationRendererOptions.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let CodeGenerationRendererOptions = /*@__PURE__*/ (() => {
    class CodeGenerationRendererOptions extends Base {
        constructor() {
            super();
            this._forceHelperLookups = false;
            this._skipAlterDataCasing = false;
            this.c = false;
            this._reusedContext = null;
            this._generateFullProject = false;
            this._forceCodeBehind = false;
            this._library = null;
            this._cDNBasePath = null;
            this._useSpacesForIndent = false;
            this._indentSpaces = 0;
            this.useSpacesForIndent = true;
            this.indentSpaces = 4;
            this.cDNBasePath = "https://static.infragistics.com/xplatform/cdn/";
        }
        get forceHelperLookups() {
            return this._forceHelperLookups;
        }
        set forceHelperLookups(a) {
            this._forceHelperLookups = a;
        }
        get skipAlterDataCasing() {
            return this._skipAlterDataCasing;
        }
        set skipAlterDataCasing(a) {
            this._skipAlterDataCasing = a;
        }
        get reusedContext() {
            return this._reusedContext;
        }
        set reusedContext(a) {
            this._reusedContext = a;
        }
        get g() {
            return this.skipAlterDataCasing || this.c;
        }
        get generateFullProject() {
            return this._generateFullProject;
        }
        set generateFullProject(a) {
            this._generateFullProject = a;
        }
        get forceCodeBehind() {
            return this._forceCodeBehind;
        }
        set forceCodeBehind(a) {
            this._forceCodeBehind = a;
        }
        get library() {
            return this._library;
        }
        set library(a) {
            this._library = a;
        }
        get cDNBasePath() {
            return this._cDNBasePath;
        }
        set cDNBasePath(a) {
            this._cDNBasePath = a;
        }
        get useSpacesForIndent() {
            return this._useSpacesForIndent;
        }
        set useSpacesForIndent(a) {
            this._useSpacesForIndent = a;
        }
        get indentSpaces() {
            return this._indentSpaces;
        }
        set indentSpaces(a) {
            this._indentSpaces = a;
        }
    }
    CodeGenerationRendererOptions.$t = /*@__PURE__*/ markType(CodeGenerationRendererOptions, 'CodeGenerationRendererOptions');
    return CodeGenerationRendererOptions;
})();
