// igniteui-react-core/esm2015/lib/CodeGeneratingComponentRendererCodeEmitter_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, NotSupportedException, IList_$type, String_$type, typeCast, Array_$type, fromEnum, Boolean_$type, markType, EnumUtil, fromEn, Number_$type, BaseError, getInstanceType, Point_$type, typeGetValue, runOn } from "./type";
import { CodeGenerationTargetPlatforms_$type } from "./CodeGenerationTargetPlatforms";
import { CodeGeneratingRendererResult } from "./CodeGeneratingRendererResult";
import { CodeGeneratingCodeWriter } from "./CodeGeneratingCodeWriter";
import { Stack$1 } from "./Stack$1";
import { CodeGenerationItemBuilderPropertyInfo } from "./CodeGenerationItemBuilderPropertyInfo";
import { List$1 } from "./List$1";
import { TypeDescriptionContext } from "./TypeDescriptionContext";
import { TypeDescriptionMetadata } from "./TypeDescriptionMetadata";
import { Rect } from "./Rect";
import { Size } from "./Size";
import { CodeGenerationSpecialValue } from "./CodeGenerationSpecialValue";
import { ColorUtil } from "./ColorUtil";
import { Brush } from "./Brush";
import { StringBuilder } from "./StringBuilder";
import { CodeGeneratingImportManager } from "./CodeGeneratingImportManager";
import { stringJoin1, stringIsNullOrEmpty, stringReplace, stringStartsWith, stringEndsWith, stringContains } from "./string";
import { HashSet$1 } from "./HashSet$1";
import { arrayGetValue } from "./array";
import { trimEnd } from "./stringExtended";
import { WebImportsHalper } from "./WebImportsHalper";
import { CodeGeneratingComponentRendererHandlerEmitter } from "./CodeGeneratingComponentRendererHandlerEmitter";
import { Dictionary$2 } from "./Dictionary$2";
import { CodeGeneratingRegexHelper } from "./CodeGeneratingRegexHelper";
import { CodeGeneratingComponentRendererDataEmitter } from "./CodeGeneratingComponentRendererDataEmitter";
import { JsonDictionaryValue } from "./JsonDictionaryValue";
import { JsonDictionaryObject } from "./JsonDictionaryObject";
import { JsonDictionaryValueType_$type } from "./JsonDictionaryValueType";
import { JsonDictionaryArray } from "./JsonDictionaryArray";
import { isNaN_ } from "./number";
import { timeSpanFromMilliseconds } from "./timespan";
import { ComponentRenderer } from "./ComponentRenderer";
import { DotNetCodeGeneratingComponentRendererHandlerEmitter } from "./DotNetCodeGeneratingComponentRendererHandlerEmitter";
import { TSCodeGeneratingComponentRendererTemplateEmitter } from "./TSCodeGeneratingComponentRendererTemplateEmitter";
import { DotNetCodeGeneratingComponentRendererTemplateEmitter } from "./DotNetCodeGeneratingComponentRendererTemplateEmitter";
import { TSCodeGeneratingComponentRendererModuleEmitter } from "./TSCodeGeneratingComponentRendererModuleEmitter";
import { DotNetCodeGeneratingComponentRendererModuleEmitter } from "./DotNetCodeGeneratingComponentRendererModuleEmitter";
import { BlazorCodeGeneratingComponentRendererModuleEmitter } from "./BlazorCodeGeneratingComponentRendererModuleEmitter";
import { CodeGenerationTransformer } from "./CodeGenerationTransformer";
import { LinearGradientBrush } from "./LinearGradientBrush";
import { IComponentRendererAdapter_$type } from "./IComponentRendererAdapter";
/**
 * @hidden
 */
export let CodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class CodeGeneratingComponentRendererCodeEmitter extends Base {
        get_al() {
            switch (this.j.platform) {
                case 1: return "Igx";
                case 2: return "Igr";
                case 3: return "Igc";
                case 4: return "Ultra";
                case 0: return "Xam";
                case 5: return "Igb";
            }
            return "";
        }
        get al() {
            return this.get_al();
        }
        get_am() {
            switch (this.j.platform) {
                case 1: return "Component";
                case 2: return "";
                case 3: return "Component";
                case 4: return "";
                case 0: return "";
            }
            return "";
        }
        get am() {
            return this.get_am();
        }
        get a6() {
            switch (this.j.platform) {
                case 1: return "angular";
                case 2: return "react";
                case 3: return "webcomponents";
            }
            return "";
        }
        get m() {
            return this.l;
        }
        br(a) {
            this.j.af(a);
        }
        get k() {
            return this.j;
        }
        constructor(a, b) {
            super();
            this.l = null;
            this.j = null;
            this.z = new List$1(CodeGenerationItemBuilderPropertyInfo.$, 0);
            this.d = null;
            this.a = null;
            this.c = null;
            this.b = null;
            this.aa = new List$1(String_$type, 0);
            this.af = new Stack$1(String_$type);
            this.ad = new Stack$1(String_$type);
            this.ae = new Stack$1(Boolean_$type);
            this.a5 = null;
            this.a4 = null;
            this.l = a;
            this.j = b;
            this.d = new CodeGeneratingCodeWriter(this.l);
            this.a = new CodeGeneratingCodeWriter(this.l);
            this.c = new CodeGeneratingCodeWriter(this.l);
            this.b = new CodeGeneratingCodeWriter(this.l);
        }
        get h() {
            return this.d;
        }
        get e() {
            return this.a;
        }
        get g() {
            return this.c;
        }
        get f() {
            return this.b;
        }
        bb(a) {
            if (a == null || a == "any") {
                return;
            }
            this.aa.add(a);
        }
        w(a, b) {
            if (this.r) {
                return false;
            }
            if (a.hasMustSetInCodePlatforms(b.b.owningType, b.b.propertyName)) {
                let c = a.getMustSetInCodePlatforms(b.b.owningType, b.b.propertyName);
                for (let e = 0; e < c.length; e++) {
                    let d = c[e];
                    let f;
                    if (((() => { let g = EnumUtil.tryParse$1(CodeGenerationTargetPlatforms_$type, d, true, f); f = g.p2; return g.ret; })())) {
                        if (f == this.j.platform) {
                            return true;
                        }
                    }
                }
            }
            return false;
        }
        q(a, b) {
            if (b.b != null && (b.b.knownType == 20 || b.b.knownType == 19 || b.b.knownType == 22 || b.b.knownType == 18)) {
                if (b.a != null) {
                    return true;
                }
            }
            if (b.b != null && (b.b.knownType == 8 || b.b.knownType == 9 || b.b.knownType == 10)) {
                if (this.w(a, b)) {
                    return true;
                }
            }
            return false;
        }
        t(a) {
            if (a.b != null && (a.b.knownType == 19)) {
                if (a.a != null) {
                    return true;
                }
            }
            return false;
        }
        get aq() {
            return this.af.d();
        }
        bc(a, b, c) {
            this.af.h(a);
        }
        u(a, b) {
            let c = false;
            if (b.getMetadata(a, "__isPlainObject") != null) {
                c = true;
            }
            return c;
        }
        bq() {
        }
        bk(a) {
        }
        bl(a) {
            this.af.e();
        }
        bi(a, b, c, d) {
        }
        bj(a, b, c, d) {
            this.bi(a, b, c, d);
        }
        toString() {
            return this.d.toString();
        }
        a0() {
            return "null";
        }
        a3(a, b, c) {
            if (b.b == null) {
                return b.c;
            }
            if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
                b.c = this.ag(b);
            }
            switch (b.b.knownType) {
                case 10:
                    if (this.w(a, b)) {
                        return this.ao(a, b);
                    }
                    return this.at(a, b);
                case 23:
                    if (this.w(a, b)) {
                        return this.ao(a, b);
                    }
                    return this.at(a, b);
                case 7: return b.c ? "true" : "false";
                case 4: return this.ai(b);
                case 6:
                    {
                        return stringJoin1(String_$type, " ", this.ab(b, c));
                    }
                case 9:
                    if (this.w(a, b)) {
                        return this.ao(a, b);
                    }
                    throw new NotSupportedException(1, "shouldn't get here");
                case 5: return this.ak(b);
                case 16:
                    {
                        return stringJoin1(String_$type, " ", this.ac(b, c));
                    }
                case 20:
                    if (b.a != null) {
                        return this.a2(b);
                    }
                    if (b.c == null) {
                        return this.a0();
                    }
                    break;
                case 22:
                    if (b.a != null) {
                        return this.a2(b);
                    }
                    if (b.c == null) {
                        return this.a0();
                    }
                    break;
                case 15: throw new NotSupportedException(1, "need to figure this out");
                case 3: throw new NotSupportedException(1, "need to figure this out");
                case 19: if (b.a != null) {
                    return this.ax(b);
                }
                else {
                    return this.a0();
                }
                case 8:
                    if (b.b.specificExternalType == "string") {
                        return this.aw(b);
                    }
                    if (b.a != null) {
                        return this.ay(b);
                    }
                    if (this.w(a, b)) {
                        return this.ao(a, b);
                    }
                    throw new NotSupportedException(1, "shouldn't get here");
                case 13: throw new NotSupportedException(1, "shouldn't get here");
                case 18: if (b.a != null) {
                    return this.az(b);
                }
                else {
                    return this.a0();
                }
                case 1: return b.c.toString();
                case 11:
                    {
                        let d = b.c;
                        return d.x + ", " + d.y;
                    }
                case 14:
                    {
                        let e = b.c;
                        return e.left + ", " + e.top + ", " + e.width + ", " + e.height;
                    }
                case 12:
                    {
                        let f = b.c;
                        return f.width + ", " + f.height;
                    }
                case 2: return b.c != null ? b.c.toString() : this.a0();
                case 21: return this.a9(b.c);
                case 17:
                    if (b.c == null) {
                        return "null";
                    }
                    if (typeof b.c === 'string') {
                        return b.c.toString();
                    }
                    return b.c.toString();
                case 0: throw new NotSupportedException(1, "shouldn't get here");
            }
            return "";
        }
        ab(a, b) {
            if (a.b == null) {
                return ((() => {
                    let $ret = new List$1(String_$type, 0);
                    $ret.add(a.c);
                    return $ret;
                })());
            }
            if (typeCast(CodeGenerationSpecialValue.$, a.c) !== null) {
                a.c = this.ag(a);
            }
            let c = a.c;
            let d = new List$1(String_$type, 0);
            for (let e = 0; e < c.count; e++) {
                d.add(this.ah(c.item(e)));
            }
            return d;
        }
        ac(a, b) {
            if (a.b == null) {
                return ((() => {
                    let $ret = new List$1(String_$type, 0);
                    $ret.add(a.c);
                    return $ret;
                })());
            }
            if (typeCast(CodeGenerationSpecialValue.$, a.c) !== null) {
                a.c = this.ag(a);
            }
            let c = a.c;
            let d = new List$1(String_$type, 0);
            for (let e = 0; e < c.count; e++) {
                d.add(this.aj(c.item(e)));
            }
            return d;
        }
        a9(a) {
            if (a == null) {
                return "";
            }
            return a.toString();
        }
        ao(a, b) {
            let c = this.a8(a, b);
            if (b.c != null) {
                let d = b.c;
                if (typeCast(List$1.$.specialize(Base.$), d) !== null) {
                    d = d.toArray();
                }
                if (typeCast(CodeGenerationItemBuilder.$, d) !== null) {
                    let e = this.i(this.l, this.j);
                    let f = this.an(c, b.c, e);
                    c = this.ar(a, b.c, f, c);
                }
                else if (typeCast(Array_$type, d) !== null) {
                    let g = d;
                    let h = new List$1(CodeGenerationItemBuilder.$, 0);
                    let i = new List$1(String_$type, 0);
                    for (let j = 0; j < g.length; j++) {
                        if (typeCast(CodeGenerationItemBuilder.$, g[j]) !== null) {
                            let k = this.i(this.l, this.j);
                            let l = c;
                            let m = g[j].p();
                            if (m == null) {
                                m = g[j].n();
                                m = this.au(m);
                            }
                            if (m != null) {
                                if (m != l) {
                                    l = m;
                                }
                                else {
                                    l = c + "_" + j.toString();
                                }
                            }
                            let n = this.an(l, g[j], k);
                            h.add(g[j]);
                            i.add(n);
                        }
                    }
                    c = this.as(b, a, h, i, c);
                }
            }
            return c;
        }
        an(a, b, c) {
            b.u = a;
            b.v(c, true);
            let d = c.toString();
            for (let e of fromEnum(c.aa)) {
                this.aa.add(e);
            }
            return d;
        }
        ar(a, b, c, d) {
            return d;
        }
        as(a, b, c, d, e) {
            return e;
        }
        a8(a, b) {
            if (b.c == null) {
                return this.a0();
            }
            if (typeCast(CodeGenerationItemBuilder.$, b.c) !== null) {
                let c = b.c.p();
                if (c != null) {
                    return c;
                }
                let d = b.c.n();
                let e = this.au(d);
                return e;
            }
            else if (typeCast(IList_$type, b.c) !== null) {
                let f = b.c;
                if (f.count > 0 && typeCast(CodeGenerationItemBuilder.$, f.item(0)) !== null) {
                    let g = f.item(0);
                    let h = g.p();
                    if (h != null) {
                        return h;
                    }
                    let i = g.n();
                    let j = this.au(i);
                    return j;
                }
                return this.au(b.b.propertyName);
            }
            else if (typeCast(Array_$type, b.c) !== null) {
                let k = b.c;
                if (k.length > 0 && typeCast(CodeGenerationItemBuilder.$, k[0]) !== null) {
                    let l = b.c.p();
                    if (l != null) {
                        return l;
                    }
                    let m = b.c.n();
                    let n = this.au(m);
                    return n;
                }
                return this.au(b.b.propertyName);
            }
            return this.a0();
        }
        at(a, b) {
            return "";
        }
        ag(a) {
            return a.c;
        }
        a2(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                if (a.a.a) {
                    return this.av(a);
                }
                else {
                    return this.a1(a);
                }
            }
            return this.a0();
        }
        s(a) {
            if (a.a != null && !stringIsNullOrEmpty(a.a.e)) {
                if (a.a.a) {
                    return true;
                }
            }
            return false;
        }
        a1(a) {
            return "";
        }
        av(a) {
            return "";
        }
        ay(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                return a.a.e;
            }
            return this.a0();
        }
        aw(a) {
            return a.c != null ? a.c.toString() : "";
        }
        az(a) {
            return this.ax(a);
        }
        ax(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                this.br(a);
                return a.a.e;
            }
            return this.a0();
        }
        get_x() {
            return false;
        }
        get x() {
            return this.get_x();
        }
        ak(a) {
            let b = a.c;
            return ColorUtil.k(b, this.x);
        }
        aj(a) {
            let b = a;
            return ColorUtil.k(b, this.x);
        }
        ai(a) {
            if (a.c == null) {
                return "";
            }
            let b = a.c;
            return ColorUtil.k(b.color, this.x);
        }
        ah(a) {
            if (a == null) {
                return "";
            }
            let b = a;
            return ColorUtil.k(b.color, this.x);
        }
        bp(a) {
            this.ad.h(a);
            this.ae.h(false);
        }
        bo(a) {
            let b = this.au(a);
            this.ad.h(b);
            this.ae.h(true);
        }
        au(a) {
            let b = stringReplace(stringReplace(a, this.a7, ""), "Component", "");
            b = CodeGeneratingComponentRenderer.e5(b);
            return this.j.v(b);
        }
        bm() {
            this.a4 = this.ad.e();
            this.ae.e();
        }
        get_a7() {
            return "";
        }
        get a7() {
            return this.get_a7();
        }
        get ap() {
            if (this.ad.f > 0) {
                return this.ad.d();
            }
            let a = stringReplace(stringReplace(this.aq, this.a7, ""), "Component", "");
            return CodeGeneratingComponentRenderer.e5(a);
        }
        get v() {
            return this.af.f == 1;
        }
        get o() {
            if (this.ae.f == 0) {
                return false;
            }
            return this.ae.d();
        }
        get p() {
            if (this.ad.f > 0) {
                return true;
            }
            return false;
        }
        get_y() {
            return true;
        }
        get y() {
            return this.get_y();
        }
        get_n() {
            return this.r;
        }
        get n() {
            return this.get_n();
        }
        bn() {
            this.a4 = this.ad.e();
        }
        be(a) {
            let b = this.f.toString();
            if (b.length > 0) {
                a.l(b);
            }
        }
        bh(a) {
            let b = this.g.toString();
            if (b.length > 0) {
                a.l(b);
            }
        }
        bd(a) {
            let b = this.e.toString();
            if (b.length > 0) {
                a.l(b);
            }
        }
        bf(a, b) {
            if (this.aa.count > 0) {
                this.bg(a, this.aa, b);
            }
        }
        bg(a, b, c) {
        }
    }
    CodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(CodeGeneratingComponentRendererCodeEmitter, 'CodeGeneratingComponentRendererCodeEmitter');
    return CodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let BlazorRazorCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class BlazorRazorCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return false;
        }
        get r() {
            return this.get_r();
        }
        ar(a, b, c, d) {
            d = super.ar(a, b, c, d);
            this.f.l("private " + b.n() + " _" + CodeGeneratingComponentRenderer.e5(d) + " = null;");
            this.f.l("public " + b.n() + " " + CodeGeneratingComponentRenderer.e6(d));
            this.f.l("{");
            this.f.h();
            this.f.l("get");
            this.f.l("{");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(d) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k(c);
            this.f.l("this._" + CodeGeneratingComponentRenderer.e5(d) + " = " + d + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(d) + ";");
            this.f.f();
            this.f.l("}");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.e6(d);
        }
        as(a, b, c, d, e) {
            e = super.as(a, b, c, d, e);
            let f = null;
            if (c.count > 0) {
                f = c._inner[0].r() + "[]";
            }
            else {
                let g = a.b.specificExternalType;
                f = CodeGenerationItemBuilder.o(this.k.platform, g, b, this.al, this.am) + "[]";
            }
            this.f.l("private " + f + " _" + CodeGeneratingComponentRenderer.e5(e) + " = null;");
            this.f.l("public " + f + " " + CodeGeneratingComponentRenderer.e6(e));
            this.f.l("{");
            this.f.h();
            this.f.l("get");
            this.f.l("{");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(e) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k("var " + CodeGeneratingComponentRenderer.e5(e) + " = new " + stringReplace(f, "[]", "") + "[" + c.count + "];");
            for (let h = 0; h < c.count; h++) {
                this.f.k(d._inner[h]);
                this.f.k(e + "[" + h + "] = " + c._inner[h].u + ";");
            }
            this.f.l("this._" + CodeGeneratingComponentRenderer.e5(e) + " = " + e + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(e) + ";");
            this.f.f();
            this.f.l("}");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.e6(e);
        }
        i(a, b) {
            return new BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        get_y() {
            return false;
        }
        constructor(a, b) {
            super(a, b);
            this.bu = 0;
            this.bs = new HashSet$1(String_$type, 0);
            this.bt = new HashSet$1(String_$type, 0);
        }
        bv(a, b, c) {
            if (c.c == null) {
                return "null";
            }
            let d = c.c;
            let e = b;
            let f = new CodeGenerationItemBuilderPropertyInfo();
            f.b = ((() => {
                let $ret = new TypeDescriptionMetadata();
                $ret.propertyName = c.b.propertyName;
                $ret.specificType = b;
                $ret.specificExternalType = b;
                return $ret;
            })());
            f.b.knownType = 8;
            f.b.owningType = c.b.owningType;
            if (f.b.specificType == "string") {
                f.b.knownType = 2;
            }
            if (f.b.specificType == "int") {
                f.b.knownType = 1;
            }
            if (f.b.specificType == "double") {
                f.b.knownType = 1;
            }
            if (f.b.knownType == 8) {
                e = this.a7 + e;
            }
            let g = "@(new " + e + "[] { ";
            for (let h = 0; h < d.length; h++) {
                if (h > 0) {
                    g += ", ";
                }
                f.c = d[h];
                if (f.b.knownType == 2) {
                    g += "\"" + this.a3(a, f, true) + "\"";
                }
                else {
                    g += this.a3(a, f, true);
                }
            }
            g += " })";
            return g;
        }
        at(a, b) {
            let c = b.b.specificType;
            if (c == null) {
                c = b.b.specificExternalType;
            }
            return this.bv(a, c, b);
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            this.h.l("<" + a + "");
            if (!this.bs.contains(this.ap)) {
                this.bs.add_1(this.ap);
                if (!this.bt.contains(this.aq)) {
                    this.bt.add_1(this.aq);
                    this.bb(this.aq);
                }
                this.f.l("private " + this.aq + " " + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                if (!this.p) {
                    if (this.v) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                }
            }
        }
        bq() {
            this.h.l(">");
            this.h.h();
        }
        bl(a) {
            this.h.f();
            this.h.l("</" + a + ">");
            this.h.l("");
            super.bl(a);
        }
        bi(a, b, c, d) {
            if (c.b != null && d.hasNameBinding(c.b.owningType, CodeGeneratingComponentRenderer.e6(c.b.propertyName)) && this.q(d, c)) {
                this.br(c);
                if (a > 0) {
                    this.h.l("");
                }
                if (c.a == null || c.a.e == null) {
                    this.h.j(b + "Name=\"@null\"");
                }
                else {
                    this.h.j(b + "Name=\"" + c.a.e + "\"");
                }
                return;
            }
            if (this.s(c)) {
                this.br(c);
                if (!this.bs.contains(this.ap)) {
                    this.bs.add_1(this.ap);
                    if (!this.bt.contains(this.aq)) {
                        this.bt.add_1(this.aq);
                        this.bb(this.aq);
                    }
                    this.f.l("private " + this.aq + " " + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                    if (!this.p) {
                        if (this.v) {
                            this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                        }
                    }
                    else {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                    }
                }
                let e = CodeGeneratingComponentRenderer.e5(this.a3(d, c, true));
                let f = e;
                if (f != "null" && !stringStartsWith(f, "this.")) {
                    f = "this." + f;
                }
                this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + "." + b + " = " + f + ";");
                return;
            }
            if (a > 0) {
                this.h.l("");
            }
            let g = false;
            if (b == "ComponentRenderer" && (c.b.owningType == "PropertyEditor" || c.b.owningType == "PropertyEditorPanel")) {
                g = true;
            }
            if (!g) {
                if (c.a != null && c.a.e != null) {
                    if (this.m != null) {
                        if (this.m.library != null) {
                            if (this.m.library.hasItem(c.a.e)) {
                                let h = this.m.library.getItem(c.a.e);
                                let i = h.getContentForPlatform(5);
                                if (i != null && i.isJSContent) {
                                    b = b + "Script";
                                }
                            }
                        }
                    }
                }
                this.h.j(b);
                this.h.j("=");
                this.h.j("\"");
                this.h.j(this.a3(d, c, true));
                this.h.j("\"");
            }
            if (b == "Name") {
                this.h.l("");
                this.h.j("@ref=\"" + CodeGeneratingComponentRenderer.e5(this.a3(d, c, true)) + "\"");
            }
        }
        aw(a) {
            let b = a.b != null ? a.b.specificType + "." : "";
            return b + (a.c != null ? a.c.toString() : "");
        }
        a3(a, b, c) {
            if (b.b == null) {
                return b.c;
            }
            if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
                b.c = this.ag(b);
            }
            switch (b.b.knownType) {
                case 11:
                    {
                        let d = b.c;
                        return "@(new Point(" + d.x + ", " + d.y + "))";
                    }
                case 14:
                    {
                        let e = b.c;
                        return "@(new Rect(" + e.left + ", " + e.top + ", " + e.width + ", " + e.height + "))";
                    }
                case 12:
                    {
                        let f = b.c;
                        return "@(new Size(" + f.width + ", " + f.height + "))";
                    }
                case 17:
                    if (b.c == null) {
                        return "null";
                    }
                    if (typeof b.c === 'string') {
                        return "@(\"" + b.c.toString() + "\")";
                    }
                    if (typeCast(Boolean_$type, b.c) !== null) {
                        return b.c ? "true" : "false";
                    }
                    return b.c.toString();
            }
            return super.a3(a, b, c);
        }
        ag(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        a0() {
            return "@null";
        }
        av(a) {
            return CodeGeneratingComponentRenderer.e5(CodeGeneratingComponentRenderer.e5(a.a.e));
        }
        bw(a) {
            a = stringReplace(a, "-", "_");
            return CodeGeneratingComponentRenderer.e6(a);
        }
        a1(a) {
            this.br(a);
            return CodeGeneratingComponentRenderer.e6(a.a.e);
        }
        bj(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.v(this, false);
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.v(this, false);
                }
            }
        }
        get_a7() {
            return "Igb";
        }
    }
    BlazorRazorCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(BlazorRazorCodeGeneratingComponentRendererCodeEmitter, 'BlazorRazorCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return BlazorRazorCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return true;
        }
        get r() {
            return this.get_r();
        }
        i(a, b) {
            return new BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        get_y() {
            return false;
        }
        constructor(a, b) {
            super(a, b);
            this.bu = 0;
            this.bs = new HashSet$1(String_$type, 0);
            this.bt = new HashSet$1(String_$type, 0);
        }
        bv(a, b) {
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "new " + a + "[] { ";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += "\"" + c[e] + "\"";
            }
            d += " }";
            return d;
        }
        at(a, b) {
            if (b.b.specificExternalType == "string") {
                return this.bv(b.b.specificExternalType, b);
            }
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d;
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            if (this.v) {
                this.h.l("var " + this.ap + " = new " + a + "();");
            }
            else {
                this.h.l("var " + this.ap + " = new " + a + "();");
            }
            if (!this.bs.contains(this.ap)) {
                this.bs.add_1(this.ap);
                if (!this.bt.contains(this.aq)) {
                    this.bt.add_1(this.aq);
                    this.bb(this.aq);
                }
                this.f.l("private " + this.aq + " " + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                if (!this.p) {
                    if (this.v) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                }
            }
        }
        bq() {
        }
        bl(a) {
            super.bl(a);
        }
        bi(a, b, c, d) {
            if (c.b != null && d.hasNameBinding(c.b.owningType, CodeGeneratingComponentRenderer.e6(c.b.propertyName)) && this.q(d, c)) {
                this.br(c);
                if (a > 0) {
                    this.h.l("");
                }
                if (c.a == null || c.a.e == null) {
                    this.h.j(this.ap + "." + b + "Name = null;");
                }
                else {
                    this.h.j(this.ap + "." + b + "Name = " + c.a.e + ";");
                }
                return;
            }
            if (this.s(c)) {
                this.br(c);
                if (!this.bs.contains(this.ap)) {
                    this.bs.add_1(this.ap);
                    if (!this.bt.contains(this.aq)) {
                        this.bt.add_1(this.aq);
                        this.bb(this.aq);
                    }
                    this.f.l("private " + this.aq + " " + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                    if (!this.p) {
                        if (this.v) {
                            this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                        }
                    }
                    else {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + ";");
                    }
                }
                let e = CodeGeneratingComponentRenderer.e5(this.a3(d, c, true));
                let f = e;
                if (f != "null" && !stringStartsWith(f, "this.")) {
                    f = "this." + f;
                }
                this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + "." + b + " = " + f + ";");
                return;
            }
            if (a > 0) {
                this.h.l("");
            }
            let g = false;
            if (b == "ComponentRenderer" && (c.b.owningType == "PropertyEditor" || c.b.owningType == "PropertyEditorPanel")) {
                g = true;
            }
            if (!g) {
                let h = false;
                if (c.a != null && c.a.e != null) {
                    if (this.m != null) {
                        if (this.m.library != null) {
                            if (this.m.library.hasItem(c.a.e)) {
                                let i = this.m.library.getItem(c.a.e);
                                let j = i.getContentForPlatform(5);
                                if (j != null && j.isJSContent) {
                                    b = b + "Script";
                                    h = true;
                                }
                            }
                        }
                    }
                }
                this.h.j(this.ap + ".");
                this.h.j(b);
                this.h.j(" = ");
                if (h) {
                    this.h.j("\"" + this.a3(d, c, true) + "\"");
                }
                else {
                    this.h.j(this.a3(d, c, true));
                }
                this.h.j(";");
            }
        }
        aw(a) {
            let b = a.b != null ? a.b.specificType + "." : "";
            return b + (a.c != null ? a.c.toString() : "");
        }
        a3(a, b, c) {
            if (b.b == null) {
                return "\"" + b.c + "\"";
            }
            if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
                b.c = this.ag(b);
            }
            switch (b.b.knownType) {
                case 11:
                    {
                        let d = b.c;
                        return "new Point(" + d.x + ", " + d.y + ")";
                    }
                case 14:
                    {
                        let e = b.c;
                        return "new Rect(" + e.left + ", " + e.top + ", " + e.width + ", " + e.height + ")";
                    }
                case 12:
                    {
                        let f = b.c;
                        return "new Size(" + f.width + ", " + f.height + ")";
                    }
                case 17:
                    if (b.c == null) {
                        return "null";
                    }
                    if (typeof b.c === 'string') {
                        return "\"" + b.c.toString() + "\"";
                    }
                    if (typeCast(Boolean_$type, b.c) !== null) {
                        return b.c ? "true" : "false";
                    }
                    return b.c.toString();
                case 5:
                case 6:
                case 16:
                case 2: return "\"" + super.a3(a, b, c) + "\"";
                case 9: throw new NotSupportedException(1, "shouldn't get here");
            }
            return super.a3(a, b, c);
        }
        ag(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        a0() {
            return "null";
        }
        av(a) {
            return CodeGeneratingComponentRenderer.e5(CodeGeneratingComponentRenderer.e5(a.a.e));
        }
        bw(a) {
            a = stringReplace(a, "-", "_");
            return CodeGeneratingComponentRenderer.e6(a);
        }
        a1(a) {
            this.br(a);
            return CodeGeneratingComponentRenderer.e6(a.a.e);
        }
        bj(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.v(this, false);
                            this.h.l(this.ap + "." + b + ".Add(" + this.a4 + ");");
                        }
                    }
                }
            }
            else if (c.b.knownType == 10) {
                let h = typeCast(Array_$type, c.c);
                if (h != null) {
                    let i = "[";
                    for (let j = 0; j < h.length; j++) {
                        let k = arrayGetValue(h, j);
                        if (typeCast(CodeGenerationItemBuilder.$, k) !== null) {
                            let l = k;
                            l.v(this, false);
                            i += this.a4 + ",";
                        }
                    }
                    i = trimEnd(i, ',');
                    i += "]";
                    this.h.l(this.ap + "." + b + " = " + i + ";");
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let m = c.c;
                    m.v(this, false);
                    this.h.l(this.ap + "." + b + " = " + this.a4 + ";");
                }
            }
        }
        get_a7() {
            return "Igb";
        }
    }
    BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let ReactMarkupCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class ReactMarkupCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return false;
        }
        get r() {
            return this.get_r();
        }
        ar(a, b, c, d) {
            d = super.ar(a, b, c, d);
            let e = b.r();
            this.f.l("private " + " _" + CodeGeneratingComponentRenderer.e5(d) + ": " + e + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.e5(d) + "(): " + e + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(d) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k(c);
            this.f.l("this._" + CodeGeneratingComponentRenderer.e5(d) + " = " + d + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(d) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.e6(d);
        }
        as(a, b, c, d, e) {
            e = super.as(a, b, c, d, e);
            let f = null;
            if (c.count > 0) {
                f = c._inner[0].r() + "[]";
            }
            else {
                let g = a.b.specificExternalType;
                f = CodeGenerationItemBuilder.o(this.k.platform, g, b, this.al, this.am) + "[]";
            }
            this.f.l("private" + " _" + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.e5(e) + "(): " + f + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(e) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k("let " + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " = [];");
            for (let h = 0; h < c.count; h++) {
                this.f.k(d._inner[h]);
                this.f.k(e + ".push(" + c._inner[h].u + ")");
            }
            this.f.l("this._" + CodeGeneratingComponentRenderer.e5(e) + " = " + e + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(e) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.e6(e);
        }
        i(a, b) {
            return new ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        constructor(a, b) {
            super(a, b);
            this.bs = new HashSet$1(String_$type, 0);
            this.bt = new HashSet$1(String_$type, 0);
        }
        at(a, b) {
            if (b.c == null) {
                return "{null}";
            }
            if (b.b != null && b.b.knownType == 23) {
                let c = b.c;
                let d = "";
                for (let e = 0; e < c.length; e++) {
                    if (e > 0) {
                        d += ", ";
                    }
                    d += c[e];
                }
                return d;
            }
            let f = b.c;
            let g = "[";
            for (let h = 0; h < f.length; h++) {
                if (h > 0) {
                    g += ", ";
                }
                if (b.b != null && (b.b.specificExternalType == "int" || b.b.specificExternalType == "double")) {
                    g += f[h].toString();
                }
                else {
                    g += "\"" + f[h].toString() + "\"";
                }
            }
            g += "]";
            return g;
        }
        get_x() {
            return true;
        }
        get_a7() {
            return "Igr";
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            this.h.l("<" + this.bv(a) + "");
            this.h.h();
            if (!this.bt.contains(this.aq)) {
                this.bt.add_1(this.aq);
                this.bb(this.aq);
            }
            if (!this.bs.contains(this.ap)) {
                this.bs.add_1(this.ap);
                this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
                if (this.v) {
                    this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + "Ref(r: " + this.aq + ") {");
                    this.f.h();
                    this.f.l("this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = r;");
                    this.f.l("this.setState({});");
                    this.f.f();
                    this.f.l("}");
                    this.g.l("this." + CodeGeneratingComponentRenderer.e5(this.ap) + "Ref = " + "this." + CodeGeneratingComponentRenderer.e5(this.ap) + "Ref.bind(this);");
                }
            }
        }
        bv(a) {
            if (stringEndsWith(a, "Component")) {
                a = a.substr(0, a.length - ("Component").length);
            }
            return CodeGeneratingComponentRenderer.e6(a);
        }
        bq() {
            if (!this.v && !this.p) {
                let a = this.bu();
                this.h.l("");
                this.h.l("name=\"" + a + "\"");
            }
            this.h.f();
            this.h.l(">");
            this.h.h();
        }
        bu() {
            let a = CodeGeneratingComponentRenderer.e5(stringReplace(stringReplace(this.aq, this.a7, ""), "Component", ""));
            a = this.k.u(a);
            return a;
        }
        ax(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                this.br(a);
                if (!this.bs.contains(a.a.e)) {
                    this.g.l("this." + CodeGeneratingComponentRenderer.e5(a.a.e) + " = " + "this." + CodeGeneratingComponentRenderer.e5(a.a.e) + ".bind(this);");
                }
                return "this." + CodeGeneratingComponentRenderer.e5(a.a.e);
            }
            return this.a0();
        }
        bl(a) {
            this.h.f();
            this.h.l("</" + this.bv(a) + ">");
            super.bl(a);
        }
        bi(a, b, c, d) {
            let e = CodeGeneratingComponentRenderer.e5(b);
            let f = this.a3(d, c, true);
            if (c.b != null && d.hasNameBinding(c.b.owningType, CodeGeneratingComponentRenderer.e6(c.b.propertyName)) && this.q(d, c)) {
                this.br(c);
                if (a > 0) {
                    this.h.l("");
                }
                if (c.a == null || c.a.e == null) {
                    this.h.j(e + "Name={null}");
                }
                else {
                    this.h.j(e + "Name=\"" + c.a.e + "\"");
                }
                return;
            }
            if (!(e == "name") || !this.v) {
                let g = (c.b != null && c.b.knownType == 10);
                if (this.q(d, c)) {
                    this.br(c);
                    let h = CodeGeneratingComponentRenderer.e5(f);
                    let i = h;
                    if (i != "null" && !stringStartsWith(i, "this.")) {
                        i = "this." + i;
                    }
                    f = "{" + i + "}";
                }
                else if (g) {
                    f = "{" + f + "}";
                }
                if (a > 0) {
                    this.h.l("");
                }
                this.h.j(e);
                this.h.j("=");
                if (!this.q(d, c) && !g) {
                    this.h.j("\"");
                }
                this.h.j(f);
                if (!this.q(d, c) && !g) {
                    this.h.j("\"");
                }
                a++;
            }
            if (e == "name") {
                if (this.v) {
                    if (a > 0) {
                        this.h.l("");
                    }
                    this.h.j("ref={this." + CodeGeneratingComponentRenderer.e5(this.a3(d, c, true)) + "Ref}");
                }
            }
        }
        ag(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        a0() {
            return "null";
        }
        av(a) {
            return "this." + CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        a1(a) {
            return "this." + CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        bj(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.v(this, false);
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.v(this, false);
                }
            }
        }
        bg(a, b, c) {
            super.bg(a, b, c);
            WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
        }
    }
    ReactMarkupCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(ReactMarkupCodeGeneratingComponentRendererCodeEmitter, 'ReactMarkupCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return ReactMarkupCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let WebCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class WebCodeBehindCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return true;
        }
        get r() {
            return this.get_r();
        }
        constructor(a, b) {
            super(a, b);
            this.bv = 0;
            this.bt = new HashSet$1(String_$type, 0);
            this.bu = new HashSet$1(String_$type, 0);
        }
        bs(a) {
            if (!a.b.owningContext.hasStringUnionPlatforms(a.b.owningType, a.b.propertyName)) {
                return false;
            }
            let b = a.b.owningContext.getStringUnionPlatforms(a.b.owningType, a.b.propertyName);
            for (let c = 0; c < b.length; c++) {
                if (b[c].toLowerCase() == this.a6) {
                    return true;
                }
            }
            return false;
        }
        aw(a) {
            if (this.bs(a)) {
                return "\"" + CodeGeneratingComponentRenderer.e7(a.c.toString()) + "\"";
            }
            else {
                if (!this.bu.contains(a.b.specificType)) {
                    this.bu.add_1(a.b.specificType);
                    if (a.b.owningContext.getMetadata(a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)) != null) {
                        let b = a.b.owningContext.getMetadata(a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)).specificExternalType;
                        this.bb(a.b.specificType + ":" + b);
                    }
                    else if (a.b.owningContext.getMetadata("Web" + a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)) != null) {
                        let c = a.b.owningContext.getMetadata("Web" + a.b.owningType, "__importTypes" + EnumUtil.getName(CodeGenerationTargetPlatforms_$type, this.k.platform)).specificExternalType;
                        this.bb(a.b.specificType + ":" + c);
                    }
                    else {
                        this.bb(a.b.specificType);
                    }
                }
                let d = a.b != null ? a.b.specificType + "." : "";
                return d + (a.c != null ? a.c.toString() : "");
            }
        }
        at(a, b) {
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "[";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d + "]";
        }
        get_x() {
            return true;
        }
        u(a, b) {
            let c = false;
            let d = this.a6 == "react" ? b.getMetadata(a, "__isPlainObject") != null && (b.getMetadata(a, "__isTSPlainInterface") != null || b.getMetadata(a, "__marshalByValue") == null) : b.getMetadata(a, "__isPlainObject") != null;
            if (d) {
                c = true;
            }
            return c;
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            let d = this.u(b, c);
            if (d) {
                if (this.v) {
                    this.h.l("var " + this.ap + ": " + a + " = {} as " + a + ";");
                }
                else {
                    this.h.l("var " + this.ap + ": " + a + " = {} as " + a + ";");
                }
            }
            else {
                if (this.v) {
                    this.h.l("var " + this.ap + " = new " + a + "();");
                }
                else {
                    this.h.l("var " + this.ap + " = new " + a + "();");
                }
            }
            if (!this.bt.contains(this.ap)) {
                this.bt.add_1(this.ap);
                if (!this.bu.contains(this.aq)) {
                    this.bu.add_1(this.aq);
                    this.bb(this.aq);
                }
                this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
                if (!this.p) {
                    if (this.v) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.querySelector('" + CodeGeneratingComponentRenderer.e8(stringReplace(this.aq, "Component", "")) + "') as " + this.aq + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.getElementById('" + this.ap + "') as " + this.aq + ";");
                }
            }
        }
        bw(a) {
            if (stringEndsWith(a, "Component")) {
                a = a.substr(0, a.length - ("Component").length);
            }
            return CodeGeneratingComponentRenderer.e8(a);
        }
        bq() {
            this.h.l("");
        }
        bk(a) {
        }
        bl(a) {
            {
                this.h.l("");
            }
            super.bl(a);
        }
        bi(a, b, c, d) {
            let e = CodeGeneratingComponentRenderer.e5(b);
            if (e == "name" && c.b == null) {
                return;
            }
            if (this.s(c)) {
                this.br(c);
                if (!this.bt.contains(this.ap)) {
                    this.bt.add_1(this.ap);
                    if (!this.bu.contains(this.aq)) {
                        this.bu.add_1(this.aq);
                        this.bb(this.aq);
                    }
                    this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
                    if (!this.p) {
                        if (this.v) {
                            this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.querySelector('" + CodeGeneratingComponentRenderer.e8(stringReplace(this.aq, "Component", "")) + "') as " + this.aq + ";");
                        }
                    }
                    else {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.getElementById('" + this.ap + "') as " + this.aq + ";");
                    }
                }
                let f = CodeGeneratingComponentRenderer.e5(this.a3(d, c, true));
                let g = f;
                if (g != "null" && !stringStartsWith(g, "this.")) {
                    g = "this." + g;
                }
                if (this.aq == "any" && stringContains(e, "-")) {
                    this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + "['" + e + "'] = " + g + ";");
                }
                else {
                    if (c.b != null && c.b.isCustomEvent) {
                        this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + ".addEventListener(\"" + e + "\", " + g + ");");
                    }
                    else {
                        this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + "." + e + " = " + g + ";");
                    }
                }
                return;
            }
            if (a > 0) {
                this.h.l("");
            }
            if (this.aq == "any" && stringContains(e, "-")) {
                this.h.j(this.ap + "." + CodeGeneratingComponentRenderer.e5(this.ap) + "['" + e + "']");
            }
            else {
                this.h.j(this.ap + "." + CodeGeneratingComponentRenderer.e5(e));
            }
            this.h.j(" = ");
            this.h.j(this.a3(d, c, true));
            this.h.j(";");
        }
        ag(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        a0() {
            return "null";
        }
        av(a) {
            return CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        a1(a) {
            return CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        ax(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                this.br(a);
                if (!this.bt.contains(a.a.e)) {
                    this.g.l("this." + CodeGeneratingComponentRenderer.e5(a.a.e) + " = " + "this." + CodeGeneratingComponentRenderer.e5(a.a.e) + ".bind(this);");
                }
                return "this." + CodeGeneratingComponentRenderer.e5(a.a.e);
            }
            return this.a0();
        }
        a3(a, b, c) {
            if (b.b == null) {
                return "\"" + b.c + "\"";
            }
            if (typeCast(CodeGenerationSpecialValue.$, b.c) !== null) {
                b.c = this.ag(b);
            }
            switch (b.b.knownType) {
                case 11:
                    {
                        let d = b.c;
                        return "{ x: " + d.x + ", y: " + d.y + "}";
                    }
                case 14:
                    {
                        let e = b.c;
                        return "{ left: " + e.left + ", top: " + e.top + ", width: " + e.width + ", height: " + e.height + "}";
                    }
                case 12:
                    {
                        let f = b.c;
                        return "{ width: " + f.width + ", height: " + f.height + "}";
                    }
                case 17:
                    if (b.c == null) {
                        return "null";
                    }
                    if (typeof b.c === 'string') {
                        return "\"" + b.c.toString() + "\"";
                    }
                    if (typeCast(Boolean_$type, b.c) !== null) {
                        return b.c ? "true" : "false";
                    }
                    return b.c.toString();
                case 5:
                case 2: return "\"" + super.a3(a, b, c) + "\"";
                case 23: return "[" + super.a3(a, b, c) + "]";
                case 6: return "[\"" + stringJoin1(String_$type, "\", \"", super.ab(b, c)) + "\"]";
                case 16: return "[\"" + stringJoin1(String_$type, "\", \"", super.ac(b, c)) + "\"]";
                case 9: throw new NotSupportedException(1, "shouldn't get here");
            }
            return super.a3(a, b, c);
        }
        bj(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    let f = true;
                    for (let g of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, g) !== null) {
                            let h = g;
                            h.v(this, false);
                            let i = c.b.owningType;
                            let j = d.getMetadata(i, "__isPlainObject");
                            let k = false;
                            if (j != null) {
                                k = true;
                            }
                            if (k) {
                                if (f) {
                                    f = false;
                                    this.h.l(this.ap + "." + b + " = []");
                                }
                                this.h.l(this.ap + "." + b + ".push(" + this.a4 + ");");
                            }
                            else {
                                this.h.l(this.ap + "." + b + ".add(" + this.a4 + ");");
                            }
                        }
                    }
                }
            }
            else if (c.b.knownType == 10) {
                let l = typeCast(Array_$type, c.c);
                if (l != null) {
                    let m = "[";
                    for (let n = 0; n < l.length; n++) {
                        let o = arrayGetValue(l, n);
                        if (typeCast(CodeGenerationItemBuilder.$, o) !== null) {
                            let p = o;
                            p.v(this, false);
                            m += this.a4 + ",";
                        }
                    }
                    m = trimEnd(m, ',');
                    m += "]";
                    this.h.l(this.ap + "." + b + " = " + m + ";");
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let q = c.c;
                    q.v(this, false);
                    this.h.l(this.ap + "." + b + " = " + this.a4 + ";");
                }
            }
        }
    }
    WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(WebCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'WebCodeBehindCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WebCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter extends WebCodeBehindCodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return true;
        }
        i(a, b) {
            return new ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        constructor(a, b) {
            super(a, b);
        }
        get_a7() {
            return "Igc";
        }
        bg(a, b, c) {
            super.bg(a, b, c);
            WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
        }
        bx(a, b) {
            let c = false;
            if (b.getMetadata(a, "__isPlainObject") != null && b.getMetadata(a, "__marshalByValue") == null) {
                c = true;
            }
            return c;
        }
    }
    ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter', WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$);
    return ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return false;
        }
        get r() {
            return this.get_r();
        }
        bs(a) {
            if (!a.b.owningContext.hasStringUnionPlatforms(a.b.owningType, a.b.propertyName)) {
                return false;
            }
            let b = a.b.owningContext.getStringUnionPlatforms(a.b.owningType, a.b.propertyName);
            for (let c = 0; c < b.length; c++) {
                if (b[c].toLowerCase() == this.a6) {
                    return true;
                }
            }
            return false;
        }
        aw(a) {
            if (this.bs(a)) {
                return CodeGeneratingComponentRenderer.e7(a.c.toString());
            }
            else {
                return super.aw(a);
            }
        }
        i(a, b) {
            return new WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        ar(a, b, c, d) {
            d = super.ar(a, b, c, d);
            let e = b.r();
            this.f.l("private" + " _" + CodeGeneratingComponentRenderer.e5(d) + ": " + e + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.e5(d) + "(): " + e + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(d) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k(c);
            this.f.l("this._" + CodeGeneratingComponentRenderer.e5(d) + " = " + d + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(d) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.e6(d);
        }
        as(a, b, c, d, e) {
            e = super.as(a, b, c, d, e);
            let f = null;
            if (c.count > 0) {
                f = c._inner[0].r() + "[]";
            }
            else {
                let g = a.b.specificExternalType;
                f = CodeGenerationItemBuilder.o(this.k.platform, g, b, this.al, this.am) + "[]";
            }
            this.f.l("private" + " _" + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.e5(e) + "(): " + f + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(e) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k("let " + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " = [];");
            for (let h = 0; h < c.count; h++) {
                this.f.k(d._inner[h]);
                this.f.k(e + ".push(" + c._inner[h].u + ")");
            }
            this.f.l("this._" + CodeGeneratingComponentRenderer.e5(e) + " = " + e + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(e) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.e6(e);
        }
        constructor(a, b) {
            super(a, b);
            this.bv = 0;
            this.bt = new HashSet$1(String_$type, 0);
            this.bu = new HashSet$1(String_$type, 0);
        }
        at(a, b) {
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d;
        }
        get_x() {
            return true;
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            this.h.l("<" + this.bw(a) + "");
            if (!this.bt.contains(this.ap)) {
                this.bt.add_1(this.ap);
                if (!this.bu.contains(this.aq)) {
                    this.bu.add_1(this.aq);
                    this.bb(this.aq);
                }
                this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
                if (!this.p) {
                    if (this.v) {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.querySelector('" + CodeGeneratingComponentRenderer.e8(stringReplace(this.aq, "Component", "")) + "') as " + this.aq + ";");
                    }
                }
                else {
                    this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.getElementById('" + this.ap + "') as " + this.aq + ";");
                }
            }
        }
        bw(a) {
            if (stringEndsWith(a, "Component")) {
                a = a.substr(0, a.length - ("Component").length);
            }
            return CodeGeneratingComponentRenderer.e8(a);
        }
        bq() {
            this.h.l(">");
            this.h.h();
        }
        bl(a) {
            this.h.f();
            this.h.l("</" + this.bw(a) + ">");
            super.bl(a);
        }
        get_a7() {
            return "Igc";
        }
        bg(a, b, c) {
            super.bg(a, b, c);
            WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
        }
        bi(a, b, c, d) {
            let e = CodeGeneratingComponentRenderer.e5(b);
            if (this.q(d, c)) {
                this.br(c);
                if (!this.bt.contains(this.ap)) {
                    this.bt.add_1(this.ap);
                    if (!this.bu.contains(this.aq)) {
                        this.bu.add_1(this.aq);
                        this.bb(this.aq);
                    }
                    this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
                    if (!this.p) {
                        if (this.v) {
                            this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.querySelector('" + CodeGeneratingComponentRenderer.e8(stringReplace(this.aq, "Component", "")) + "') as " + this.aq + ";");
                        }
                    }
                    else {
                        this.g.l("var " + CodeGeneratingComponentRenderer.e5(this.ap) + " = this." + CodeGeneratingComponentRenderer.e5(this.ap) + " = document.getElementById('" + this.ap + "') as " + this.aq + ";");
                    }
                }
                let f = CodeGeneratingComponentRenderer.e5(this.a3(d, c, true));
                let g = f;
                if (g != "null" && !stringStartsWith(g, "this.")) {
                    g = "this." + g;
                }
                if (c.b.isCustomEvent) {
                    this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + ".addEventListener(\"" + e + "\", " + g + ");");
                }
                else {
                    this.e.l(CodeGeneratingComponentRenderer.e5(this.ap) + "." + e + " = " + g + ";");
                }
                return;
            }
            if (a > 0) {
                this.h.l("");
            }
            this.h.j(CodeGeneratingComponentRenderer.e8(e));
            this.h.j("=");
            this.h.j("\"");
            this.h.j(this.a3(d, c, true));
            this.h.j("\"");
            if (e == "name") {
                this.h.l("");
                this.h.j("id=\"" + this.a3(d, c, true) + "\"");
            }
        }
        ag(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        a0() {
            return "null";
        }
        av(a) {
            return CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        a1(a) {
            return CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        ax(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                this.br(a);
                if (!this.bt.contains(a.a.e)) {
                    this.g.l("this." + CodeGeneratingComponentRenderer.e5(a.a.e) + " = " + "this." + CodeGeneratingComponentRenderer.e5(a.a.e) + ".bind(this);");
                }
                return "this." + CodeGeneratingComponentRenderer.e5(a.a.e);
            }
            return this.a0();
        }
        bj(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.v(this, false);
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.v(this, false);
                }
            }
        }
    }
    WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter, 'WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter extends WebCodeBehindCodeGeneratingComponentRendererCodeEmitter {
        i(a, b) {
            return new WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        constructor(a, b) {
            super(a, b);
        }
        get_a7() {
            return "Igc";
        }
        bg(a, b, c) {
            super.bg(a, b, c);
            WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
        }
    }
    WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter', WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$);
    return WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let TSCodeGeneratingComponentRendererHandlerEmitter = /*@__PURE__*/ (() => {
    class TSCodeGeneratingComponentRendererHandlerEmitter extends CodeGeneratingComponentRendererHandlerEmitter {
        constructor(a, b, c) {
            super(a, b);
            this.w = 0;
            this.x = new Dictionary$2(String_$type, String_$type, 0);
            this.w = c;
            this.x.item("IgxRowType", "RowType");
            this.x.item("IgxCellType", "CellType");
        }
        t(a, b) {
            super.t(a, b);
            this.g.l("this." + CodeGeneratingComponentRenderer.e5(a) + "();");
        }
        u(a, b) {
            super.u(a, b);
            this.h.l("this." + CodeGeneratingComponentRenderer.e5(a) + "();");
        }
        get aa() {
            switch (this.w) {
                case 1: return "angular";
                case 2: return "react";
                case 3: return "webcomponents";
            }
            return "";
        }
        get ab() {
            switch (this.w) {
                case 1: return "Igx";
                case 2: return "Igr";
                case 3: return "Igc";
            }
            return "";
        }
        v(a, b) {
            super.v(a, b);
            let c = new CodeGeneratingRegexHelper("import {([^}]+)}\\s+from\\s+['\"]([^'\"]+)['\"]");
            c.execute(b, (d, e, f) => {
                let g = d(1);
                let h = d(2);
                h = stringReplace(h, "webcomponents", this.aa);
                h = stringReplace(h, "react", this.aa);
                h = stringReplace(h, "angular", this.aa);
                if (this.aa == "angular" || this.aa == "react") {
                    if (stringContains(h, "/grids")) {
                        h = stringReplace(h, "/grids", "");
                    }
                }
                if (this.aa == "angular") {
                    if (stringContains(h, "igniteui-angular-grids")) {
                        h = stringReplace(h, "igniteui-angular-grids", "igniteui-angular");
                    }
                }
                let i = g.split(',');
                for (let k = 0; k < i.length; k++) {
                    let j = i[k];
                    let l = j.trim();
                    if (stringStartsWith(l, "Igc")) {
                        l = this.ab + l.substr(3);
                    }
                    if (stringStartsWith(l, "Igr")) {
                        l = this.ab + l.substr(3);
                    }
                    if (stringStartsWith(l, "Igx")) {
                        l = this.ab + l.substr(3);
                    }
                    if (stringStartsWith(l, "Igr") && stringEndsWith(l, "Component")) {
                        l = l.substr(0, l.length - ("Component").length);
                    }
                    if (h.trim() == "igniteui-angular") {
                        if (this.x.containsKey(l.trim())) {
                            l = this.x.item(l.trim());
                        }
                        if (stringStartsWith(l.trim(), "Igx") && stringEndsWith(l.trim(), "EventArgs")) {
                            this.x.item(l.trim(), stringReplace(l, "Igx", "I"));
                            l = stringReplace(l, "Igx", "I");
                        }
                    }
                    a.addPackageImport(h.trim(), l.trim());
                }
                return null;
            });
        }
        get_z() {
            switch (this.n.platform) {
                case 1: return "Component";
                case 2: return "";
                case 3: return "Component";
                case 4: return "";
                case 0: return "";
            }
            return "";
        }
        get z() {
            return this.get_z();
        }
        get_y() {
            switch (this.n.platform) {
                case 1: return "Igx";
                case 2: return "Igr";
                case 3: return "Igc";
                case 4: return "Ultra";
                case 0: return "Xam";
                case 5: return "Igb";
            }
            return "";
        }
        get y() {
            return this.get_y();
        }
        q(a, b) {
            return a;
        }
    }
    TSCodeGeneratingComponentRendererHandlerEmitter.$t = /*@__PURE__*/ markType(TSCodeGeneratingComponentRendererHandlerEmitter, 'TSCodeGeneratingComponentRendererHandlerEmitter', CodeGeneratingComponentRendererHandlerEmitter.$);
    return TSCodeGeneratingComponentRendererHandlerEmitter;
})();
/**
 * @hidden
 */
export let TSCodeGeneratingComponentRendererDataEmitter = /*@__PURE__*/ (() => {
    class TSCodeGeneratingComponentRendererDataEmitter extends CodeGeneratingComponentRendererDataEmitter {
        constructor(a, b) {
            super(a, b);
            this.am = new Dictionary$2(CodeGeneratingCodeWriter.$, Number_$type, 0);
        }
        x(a, b = false) {
            a = CodeGeneratingComponentRenderer.e6(a);
            a = stringReplace(a, "-", "_");
            if (b) {
                a += "Item";
            }
            return a;
        }
        w(a) {
            a = stringReplace(a, "-", "_");
            if (this.p.g) {
                return a;
            }
            return CodeGeneratingComponentRenderer.e5(a);
        }
        aq(a) {
            a = stringReplace(a, "-", "_");
            return "_" + CodeGeneratingComponentRenderer.e5(a);
        }
        get_t() {
            return ".ts";
        }
        al(a, b, c, d, e, f) {
            super.al(a, b, c, d, e, f);
            let g = this.s(this.x(b, true), this.x(b, false), c, d);
            for (let h of fromEnum(g)) {
                this.ad(h, e, f);
            }
            let i = 0;
            if (f == this.k) {
                f.addPackageImport("./" + this.x(b, false), this.x(b, false));
            }
            else {
                f.addPackageImport("./" + this.x(b, false), this.x(b, false));
            }
            e.l("export class " + this.x(b, false) + " extends Array<" + this.x(b, true) + "> {");
            e.h();
            e.l("public constructor(items: Array<" + this.x(b, true) + "> | number = -1) {");
            e.h();
            e.l("if (Array.isArray(items)) {");
            e.h();
            e.l("super(...items);");
            e.f();
            e.l("} else {");
            e.h();
            e.l("const newItems = [");
            e.h();
        }
        ah(a, b, c) {
            super.ah(a, b, c);
            let d = this.p.cDNBasePath + this.x(a, false) + ".json";
            let e = "";
            if (this.n.platform == 2) {
                e = "this.setState({}); ";
            }
            if (this.n.platform == 3) {
                e = "if ((this as any)._bind) { (this as any)._bind(); } ";
            }
            if (this.n.platform == 1) {
                e = "this._detector.markForCheck(); ";
            }
            let f = this.x(a, false);
            let g = "null";
            switch (b) {
                case 3:
                case 2:
                    f = "any[]";
                    break;
                case 1:
                    f = "any[]";
                    g = this.x(a, false);
                    break;
                case 0: break;
            }
            this.g.l("private " + this.aq(a) + ": " + f + " = " + g + ";");
            if ((b == 3 || b == 2) || (c != null && c.isAsync)) {
                this.g.l("private _isFetching: boolean = false;");
            }
            this.g.l("public get " + CodeGeneratingComponentRenderer.e5(a) + "(): " + f + " {");
            this.g.h();
            if (c != null && c.isAsync) {
                this.g.l("if (this." + this.aq(a) + " == null && !this._isFetching)");
                this.g.l("{");
                this.g.h();
                this.g.l("this._isFetching = true;");
                this.g.l("( async () => { this." + this.aq(a) + " = await (await " + this.x(a, false) + ".fetch()); " + e + " })();");
                this.g.f();
                this.g.l("}");
            }
            else {
                if ((b == 3 || b == 2)) {
                    this.g.l("if (this." + this.aq(a) + " == null && !this._isFetching)");
                    this.g.l("{");
                    this.g.h();
                    this.g.l("this._isFetching = true;");
                    this.g.l("( async () => {");
                    this.g.l("    const url = '" + d + "';");
                    this.g.l("    this." + this.aq(a) + " = await (await fetch(url)).json();");
                    this.g.l("    " + e);
                    this.g.l("})();");
                    this.g.f();
                    this.g.l("}");
                }
                if (b == 0) {
                    this.g.l("if (this." + this.aq(a) + " == null)");
                    this.g.l("{");
                    this.g.h();
                    this.g.l("this." + this.aq(a) + " = new " + this.x(a, false) + "();");
                    this.g.f();
                    this.g.l("}");
                }
            }
            this.g.l("return this." + this.aq(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        ai(a, b) {
            super.ai(a, b);
            this.g.l("private " + this.aq(a) + ": " + this.an(b.e) + " = " + this.ao(b) + ";");
            this.g.l("public get " + CodeGeneratingComponentRenderer.e5(a) + "(): " + this.an(b.e) + " {");
            this.g.h();
            this.g.l("return this." + this.aq(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        aj(a, b, c, d) {
            super.aj(a, b, c, d);
            this.g.l("private " + this.aq(a) + ": " + this.x(a, false) + "Item = null;");
            this.g.l("public get " + CodeGeneratingComponentRenderer.e5(a) + "(): " + this.x(a, false) + "Item {");
            this.g.h();
            this.g.l("if (this." + this.aq(a) + " == null)");
            this.g.l("{");
            this.g.h();
            this.g.l("this." + this.aq(a) + " = ");
            this.ar(this.x(a, true), b, c, d, this.g);
            this.g.l("}");
            this.g.l("return this." + this.aq(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        z(a, b) {
            b.addPackageImport("./" + a, a);
        }
        aa(a, b, c) {
            c.addDefaultImport("./" + b, a);
        }
        ab(a, b) {
            this.l = CodeGeneratingImportManager.merge(this.l, b);
        }
        ad(a, b, c) {
            let d = 0;
            if (c == this.k) {
                c.addPackageImport("./" + a.c, a.d);
            }
            else {
                c.addPackageImport("./" + a.c, a.d);
            }
            b.l("export class " + a.d + " {");
            b.h();
            b.l("public constructor(init: Partial<" + a.d + ">) {");
            b.h();
            b.l("Object.assign(this, init);");
            b.f();
            b.l("}");
            b.l("");
            for (let e = 0; e < a.a.length; e++) {
                let f = a.a[e];
                let g = a.b[e];
                b.l("public " + this.w(f) + ": " + this.ap(a, f, g) + ";");
            }
            b.f();
            b.l("");
            b.l("}");
        }
        an(a) {
            switch (a) {
                case 1: return "boolean";
                case 0: return "number";
                case 2: return "string";
                case 3: return "any";
            }
            return "any";
        }
        ap(a, b, c) {
            if (c.d) {
                return a.d + "_" + CodeGeneratingComponentRenderer.e6(this.w(b)) + "Item[]";
            }
            else if (c.f) {
                return a.d + "_" + TypeDescriptionContext.ab(this.w(b));
            }
            else {
                switch (c.c) {
                    case 1: return "boolean";
                    case 0: return "number";
                    case 2: return "string";
                    case 3: return "any";
                }
            }
            return "any";
        }
        ak(a, b, c, d, e) {
            super.ak(a, b, c, d, e);
            e.f();
            e.l("];");
            e.l("super(...newItems.slice(0));");
            e.f();
            e.l("}");
            e.f();
            e.l("}");
            e.f();
            e.l("}");
        }
        y(a, b, c, d, e, f, g) {
            super.y(a, b, c, d, e, f, g);
            this.ar(this.x(b, true), c, d, e, g);
            g.l(",");
        }
        ar(a, b, c, d, e) {
            if (d == null) {
                e.j("null");
                return;
            }
            if (typeCast(JsonDictionaryArray.$, d) !== null) {
                let f = d;
                e.l("[");
                e.h();
                if (f.items != null && f.items.length > 0) {
                    for (let g = 0; g < f.items.length; g++) {
                        if (g > 0) {
                            e.l(",");
                        }
                        this.ar(a, b, c, f.items[g], e);
                    }
                }
                e.f();
                e.l("]");
            }
            else if (typeCast(JsonDictionaryObject.$, d) !== null) {
                let h = d;
                e.l("new " + a + "(");
                e.l("{");
                e.h();
                let i = true;
                for (let j = 0; j < b.length; j++) {
                    let k = b[j];
                    let l = c[j];
                    if (!h.g(k)) {
                        continue;
                    }
                    if (i) {
                        i = false;
                    }
                    else {
                        e.l(",");
                    }
                    e.j(this.w(k) + ": ");
                    if (l.d) {
                        let m = a + "_" + CodeGeneratingComponentRenderer.e6(this.w(k)) + "Item";
                        if (h.g(k)) {
                            this.ar(m, l.a, l.b, h.item(k), e);
                        }
                        else {
                            e.j("[]");
                        }
                    }
                    else if (l.f) {
                        let n = a + "_" + CodeGeneratingComponentRenderer.e6(this.w(k));
                        this.ar(n, l.a, l.b, h.item(k), e);
                    }
                    else {
                        let o = h.item(k);
                        e.j(this.u(l, o));
                    }
                }
                e.l("");
                e.f();
                e.j("})");
            }
        }
        u(a, b) {
            if (b.value == null) {
                return ("null");
            }
            switch (a.c) {
                case 1: return (b.value ? "true" : "false");
                case 0: return (b.value.toString());
                case 2: return ("`" + stringReplace(b.value, "`", "\\`") + "`");
                case 3: return ("null");
            }
            return "";
        }
        ao(a) {
            switch (a.e) {
                case 1: return (a.value ? "true" : "false");
                case 0: return (a.value.toString());
                case 2: return ("`" + a.value + "`");
                case 3: return "null";
            }
            return "";
        }
    }
    TSCodeGeneratingComponentRendererDataEmitter.$t = /*@__PURE__*/ markType(TSCodeGeneratingComponentRendererDataEmitter, 'TSCodeGeneratingComponentRendererDataEmitter', CodeGeneratingComponentRendererDataEmitter.$);
    return TSCodeGeneratingComponentRendererDataEmitter;
})();
/**
 * @hidden
 */
export let AngularTemplateCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class AngularTemplateCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return false;
        }
        get r() {
            return this.get_r();
        }
        ar(a, b, c, d) {
            d = super.ar(a, b, c, d);
            let e = b.r();
            this.f.l("private " + " _" + CodeGeneratingComponentRenderer.e5(d) + ": " + e + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.e5(d) + "(): " + e + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(d) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k(c);
            this.f.l("this._" + CodeGeneratingComponentRenderer.e5(d) + " = " + d + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(d) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.e5(d);
        }
        as(a, b, c, d, e) {
            e = super.as(a, b, c, d, e);
            let f = null;
            if (c.count > 0) {
                f = c._inner[0].r() + "[]";
            }
            else {
                let g = a.b.specificExternalType;
                f = CodeGenerationItemBuilder.o(this.k.platform, g, b, this.al, this.am) + "[]";
            }
            this.f.l("private" + " _" + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " | null = null;");
            this.f.l("public get " + CodeGeneratingComponentRenderer.e5(e) + "(): " + f + " {");
            this.f.h();
            this.f.l("if (this._" + CodeGeneratingComponentRenderer.e5(e) + " == null)");
            this.f.l("{");
            this.f.h();
            this.f.k("let " + CodeGeneratingComponentRenderer.e5(e) + ": " + f + " = [];");
            for (let h = 0; h < c.count; h++) {
                this.f.k(d._inner[h]);
                this.f.k(e + ".push(" + c._inner[h].u + ")");
            }
            this.f.l("this._" + CodeGeneratingComponentRenderer.e5(e) + " = " + e + ";");
            this.f.f();
            this.f.l("}");
            this.f.l("return this._" + CodeGeneratingComponentRenderer.e5(e) + ";");
            this.f.f();
            this.f.l("}");
            return CodeGeneratingComponentRenderer.e6(e);
        }
        i(a, b) {
            return new AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        constructor(a, b) {
            super(a, b);
            this.bu = 0;
            this.bs = new HashSet$1(String_$type, 0);
            this.bt = new HashSet$1(String_$type, 0);
        }
        at(a, b) {
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d;
        }
        get_x() {
            return true;
        }
        get_a7() {
            return "Igx";
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            this.h.l("<" + this.bv(a) + "");
            if (!this.bs.contains(this.ap)) {
                this.bs.add_1(this.ap);
                if (!this.bt.contains(this.aq)) {
                    this.bt.add_1(this.aq);
                    this.bb(this.aq);
                }
                this.f.l("@ViewChild(\"" + CodeGeneratingComponentRenderer.e5(this.ap) + "\", { static: true } )");
                this.f.l("private " + CodeGeneratingComponentRenderer.e5(this.ap) + ": " + this.aq);
            }
        }
        bg(a, b, c) {
            super.bg(a, b, c);
            WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
        }
        bv(a) {
            if (stringEndsWith(a, "Component")) {
                a = a.substr(0, a.length - ("Component").length);
            }
            return CodeGeneratingComponentRenderer.e8(a);
        }
        bq() {
            this.h.l(">");
            this.h.h();
        }
        bl(a) {
            this.h.f();
            this.h.l("</" + this.bv(a) + ">");
            super.bl(a);
        }
        bi(a, b, c, d) {
            let e = CodeGeneratingComponentRenderer.e5(b);
            if (this.q(d, c)) {
                this.br(c);
                if (this.t(c)) {
                    e = "(" + e + ")";
                }
                else {
                    e = "[" + e + "]";
                }
            }
            if (a > 0) {
                this.h.l("");
            }
            this.h.j(e);
            this.h.j("=");
            this.h.j("\"");
            this.h.j(this.a3(d, c, true));
            this.h.j("\"");
            if (e == "name") {
                this.h.l("");
                this.h.j("#" + CodeGeneratingComponentRenderer.e5(this.a3(d, c, true)));
            }
        }
        ax(a) {
            if (!stringIsNullOrEmpty(a.a.e)) {
                this.br(a);
                return "this." + CodeGeneratingComponentRenderer.e5(a.a.e) + "($event)";
            }
            return this.a0();
        }
        ag(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        a0() {
            return "null";
        }
        av(a) {
            return CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        a1(a) {
            return CodeGeneratingComponentRenderer.e5(a.a.e);
        }
        bj(a, b, c, d) {
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.v(this, false);
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.v(this, false);
                }
            }
        }
    }
    AngularTemplateCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(AngularTemplateCodeGeneratingComponentRendererCodeEmitter, 'AngularTemplateCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return AngularTemplateCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter extends WebCodeBehindCodeGeneratingComponentRendererCodeEmitter {
        get_r() {
            return true;
        }
        i(a, b) {
            return new AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        constructor(a, b) {
            super(a, b);
        }
        get_a7() {
            return "Igc";
        }
        bg(a, b, c) {
            super.bg(a, b, c);
            WebImportsHalper.b(this.k, this.a6, this.a5, this.a7, a, b, c);
        }
    }
    AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter', WebCodeBehindCodeGeneratingComponentRendererCodeEmitter.$);
    return AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let DotNetCodeGeneratingComponentRendererDataEmitter = /*@__PURE__*/ (() => {
    class DotNetCodeGeneratingComponentRendererDataEmitter extends CodeGeneratingComponentRendererDataEmitter {
        constructor(a, b) {
            super(a, b);
            this.am = false;
        }
        ab(a, b) {
            super.ab(a, b);
            this.f.l("<Compile Include=\"" + a + ".cs\">");
            this.f.h();
            this.f.l("<SubType>Code</SubType>");
            this.f.f();
            this.f.l("</Compile>");
        }
        af(a) {
            a.l("using System;");
            a.l("using System.Collections.Generic;");
            if (!this.am) {
                this.am = true;
            }
            super.af(a);
        }
        ae(a) {
            super.ae(a);
        }
        x(a, b = false) {
            a = CodeGeneratingComponentRenderer.e6(a);
            a = stringReplace(a, "-", "_");
            if (b) {
                a += "Item";
            }
            return a;
        }
        w(a) {
            a = stringReplace(a, "-", "_");
            return CodeGeneratingComponentRenderer.e6(a);
        }
        aq(a) {
            a = stringReplace(a, "-", "_");
            return "_" + CodeGeneratingComponentRenderer.e5(a);
        }
        al(a, b, c, d, e, f) {
            super.al(a, b, c, d, e, f);
            let g = this.s(this.x(b, true), this.x(b, false), c, d);
            for (let h of fromEnum(g)) {
                this.ad(h, e, f);
            }
            e.l("");
            e.l("public class " + this.x(b, false));
            e.h();
            e.l(": List<" + this.x(b, true) + ">");
            e.f();
            e.l("{");
            e.h();
            e.l("public " + this.x(b, false) + "()");
            e.l("{");
            e.h();
        }
        ah(a, b, c) {
            super.ah(a, b, c);
            if (b == 3) {
                let d = this.p.cDNBasePath + this.x(a, false) + ".json";
                this.g.l("public RemoteJSON " + this.w(a));
                this.g.l("{");
                this.g.h();
                this.g.l("get");
                this.g.l("{");
                this.g.h();
                this.g.l("return RemoteJSON.LoadFrom(\"" + d + "\");");
                this.g.f();
                this.g.l("}");
                this.g.l("");
                return;
            }
            let e = this.aq(a) + " = " + "new " + this.x(a, false) + "(); ";
            let f = "";
            if (this.n.platform == 5) {
                f = " StateHasChanged();";
            }
            else {
                f = " OnPropertyChanged(\"" + this.x(a, false) + "\");";
            }
            if (c != null && (b == 2 || c.isAsync)) {
                e = this.x(a, false) + ".Fetch().ContinueWith((t) => {" + this.aq(a) + " = t.Result; " + f + " });";
            }
            this.g.l("private " + this.x(a, false) + " " + this.aq(a) + " = null;");
            this.g.l("public " + this.x(a, false) + " " + this.w(a));
            this.g.l("{");
            this.g.h();
            this.g.l("get");
            this.g.l("{");
            this.g.h();
            this.g.l("if (" + this.aq(a) + " == null)");
            this.g.l("{");
            this.g.h();
            this.g.l(e);
            this.g.f();
            this.g.l("}");
            this.g.l("return " + this.aq(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        ai(a, b) {
            super.ai(a, b);
            this.g.l("private " + this.an(b.e) + " " + this.aq(a) + " = " + this.ao(b) + ";");
            this.g.l("public " + this.an(b.e) + " " + this.w(a));
            this.g.l("{");
            this.g.h();
            this.g.l("get");
            this.g.l("{");
            this.g.h();
            this.g.l("return " + this.aq(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        aj(a, b, c, d) {
            super.aj(a, b, c, d);
            this.g.l("private " + this.x(a, false) + "Item " + this.aq(a) + " = null;");
            this.g.l("public " + this.x(a, false) + "Item " + this.w(a));
            this.g.l("{");
            this.g.h();
            this.g.l("get");
            this.g.l("{");
            this.g.h();
            this.g.l("if (" + this.aq(a) + " == null)");
            this.g.l("{");
            this.g.h();
            this.g.l("_" + a + " = ");
            this.ar(this.x(a, true), b, c, d, this.g);
            this.g.f();
            this.g.l("}");
            this.g.l("return " + this.aq(a) + ";");
            this.g.f();
            this.g.l("}");
            this.g.f();
            this.g.l("}");
            this.g.l("");
        }
        get_t() {
            return ".cs";
        }
        ad(a, b, c) {
            b.l("public class " + a.d);
            b.l("{");
            b.h();
            for (let d = 0; d < a.a.length; d++) {
                let e = a.a[d];
                let f = a.b[d];
                b.l("public " + this.ap(a, e, f) + " " + this.w(e) + " { get; set; }");
            }
            b.f();
            b.l("}");
        }
        an(a) {
            switch (a) {
                case 1: return "bool";
                case 0: return "double";
                case 2: return "string";
                case 3: return "object";
            }
            return "object";
        }
        get_r() {
            return true;
        }
        ap(a, b, c) {
            if (c.d) {
                return "List<" + a.d + "_" + this.w(b) + "Item>";
            }
            else if (c.f) {
                return a.d + "_" + this.w(b);
            }
            else {
                switch (c.c) {
                    case 1: return "bool";
                    case 0: return "double";
                    case 2: return "string";
                    case 3: return "object";
                }
            }
            return "object";
        }
        ak(a, b, c, d, e) {
            super.ak(a, b, c, d, e);
            e.f();
            e.l("}");
            if (a == 2 || a == 3) {
                let f = this.p.cDNBasePath + this.x(b, false) + ".json";
                e.l("public async static System.Threading.Tasks.Task<" + this.x(b, false) + "> Fetch()");
                e.l("{");
                e.h();
                e.l("var url = \"" + f + "\" ;");
                e.l("var client = new System.Net.Http.HttpClient();");
                e.l("var str = await client.GetStringAsync(url);");
                e.l("var arr = System.Text.Json.JsonSerializer.Deserialize<" + this.x(b, false) + ">(str);");
                e.l("return arr;");
                e.f();
                e.l("}");
                e.l("");
            }
            e.f();
            e.l("}");
        }
        y(a, b, c, d, e, f, g) {
            super.y(a, b, c, d, e, f, g);
            if (a == 2 || a == 3) {
                return;
            }
            g.j("this.Add(");
            this.ar(this.x(b, true), c, d, e, g);
            g.l(");");
        }
        ar(a, b, c, d, e) {
            if (d == null) {
                e.j("null");
                return;
            }
            if (typeCast(JsonDictionaryArray.$, d) !== null) {
                let f = d;
                e.l("new List<" + a + ">()");
                e.l("{");
                e.h();
                if (f.items != null && f.items.length > 0) {
                    for (let g = 0; g < f.items.length; g++) {
                        if (g > 0) {
                            e.l(",");
                        }
                        this.ar(a, b, c, f.items[g], e);
                    }
                }
                e.f();
                e.l("}");
            }
            else if (typeCast(JsonDictionaryObject.$, d) !== null) {
                let h = d;
                e.l("new " + a + "()");
                e.l("{");
                e.h();
                for (let i = 0; i < b.length; i++) {
                    let j = b[i];
                    let k = c[i];
                    if (!h.g(j)) {
                        continue;
                    }
                    if (i > 0) {
                        e.l(",");
                    }
                    e.j(j + " = ");
                    if (k.d) {
                        let l = a + "_" + this.w(j) + "Item";
                        this.ar(l, k.a, k.b, h.item(j), e);
                    }
                    else if (k.f) {
                        let m = a + "_" + this.w(j);
                        this.ar(m, k.a, k.b, h.item(j), e);
                    }
                    else {
                        let n = h.item(j);
                        try {
                            e.j(this.u(k, n));
                        }
                        catch (o) {
                            let p = getInstanceType(n.value).typeName;
                            let q = stringReplace(EnumUtil.getName(JsonDictionaryValueType_$type, k.c), "Value", "");
                            let r = "'" + j + "' property (" + n.value + ") is '" + p + "' instead of '" + q + "' type.";
                            throw new BaseError(1, "Failed on creating '" + a + "' because " + r + "\n" + o.message);
                        }
                    }
                }
                e.l("");
                e.f();
                e.j("}");
            }
        }
        u(a, b) {
            switch (a.c) {
                case 1: return (b.value ? "true" : "false");
                case 0:
                    if (b == null || b.value == null) {
                        return "double.NaN";
                    }
                    return (b.value.toString());
                case 2: if (b.value == null) {
                    return "null";
                }
                else {
                    return ("@\"" + stringReplace(b.value, "\"", "\"\"") + "\"");
                }
                case 3: return ("null");
            }
            return "";
        }
        ao(a) {
            switch (a.e) {
                case 1: return (a.value ? "true" : "false");
                case 0: return (a.value.toString());
                case 2: return ("@\"" + stringReplace(a.value, "\"", "\"\"") + "\"");
                case 3: return "null";
            }
            return "";
        }
    }
    DotNetCodeGeneratingComponentRendererDataEmitter.$t = /*@__PURE__*/ markType(DotNetCodeGeneratingComponentRendererDataEmitter, 'DotNetCodeGeneratingComponentRendererDataEmitter', CodeGeneratingComponentRendererDataEmitter.$);
    return DotNetCodeGeneratingComponentRendererDataEmitter;
})();
/**
 * @hidden
 */
export let WPFXamlCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class WPFXamlCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        a9(a) {
            if (a == null || isNaN_(a)) {
                return "";
            }
            return timeSpanFromMilliseconds(a).toString();
        }
        get_r() {
            return false;
        }
        get r() {
            return this.get_r();
        }
        i(a, b) {
            return new WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        get_y() {
            return false;
        }
        constructor(a, b) {
            super(a, b);
        }
        at(a, b) {
            if (b.c == null) {
                return "null";
            }
            let c = b.c;
            let d = "";
            for (let e = 0; e < c.length; e++) {
                if (e > 0) {
                    d += ", ";
                }
                d += c[e];
            }
            return d;
        }
        bc(a, b, c) {
            super.bc(a, b, c);
            this.h.l("<ig:" + a + "");
        }
        bq() {
            this.h.l(">");
            this.h.h();
        }
        bl(a) {
            this.h.f();
            this.h.l("</ig:" + a + ">");
            super.bl(a);
        }
        bi(a, b, c, d) {
            if (b == "Label" && c.b != null && stringContains(c.b.owningType, "Axis")) {
                if (c.c != null && stringContains(c.c, "{")) {
                    if (a > 0) {
                        this.h.l("");
                    }
                    this.h.j(b);
                    this.h.j("=");
                    this.h.j("\"");
                    this.h.j("{}" + this.a3(d, c, true));
                    this.h.j("\"");
                    return;
                }
            }
            if (a > 0) {
                this.h.l("");
            }
            this.h.j(b);
            this.h.j("=");
            this.h.j("\"");
            this.h.j(this.a3(d, c, true));
            this.h.j("\"");
        }
        ag(a) {
            let b = a.c;
            switch (b.a) {
                case 3: return b.b;
                case 0: return b.b;
                case 1: return b.b;
                case 2: return b.b;
            }
            return b.b;
        }
        a0() {
            return "{x:Null}";
        }
        av(a) {
            return "{Binding ElementName=" + a.a.e + "}";
        }
        bs(a) {
            a = stringReplace(a, "-", "_");
            return CodeGeneratingComponentRenderer.e6(a);
        }
        a1(a) {
            this.br(a);
            return "{Binding " + this.bs(a.a.e) + "}";
        }
        bj(a, b, c, d) {
            this.h.j("<ig:");
            {
                this.h.j(this.aq);
                this.h.j(".");
            }
            this.h.j(b);
            this.h.l(">");
            this.h.h();
            if (c.b.knownType == 9) {
                let e = typeCast(IList_$type, c.c);
                if (e != null) {
                    for (let f of fromEn(e)) {
                        if (typeCast(CodeGenerationItemBuilder.$, f) !== null) {
                            let g = f;
                            g.v(this, false);
                        }
                    }
                }
            }
            else {
                if (typeCast(CodeGenerationItemBuilder.$, c.c) !== null) {
                    let h = c.c;
                    h.v(this, false);
                }
            }
            this.h.f();
            this.h.j("</ig:");
            {
                this.h.j(this.aq);
                this.h.j(".");
            }
            this.h.j(b);
            this.h.l(">");
        }
        get_a7() {
            return "Xam";
        }
    }
    WPFXamlCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(WPFXamlCodeGeneratingComponentRendererCodeEmitter, 'WPFXamlCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WPFXamlCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter = /*@__PURE__*/ (() => {
    class WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter extends CodeGeneratingComponentRendererCodeEmitter {
        a9(a) {
            if (a == null || isNaN_(a)) {
                return "";
            }
            return "TimeSpan.FromMilliseconds(" + a + ")";
        }
        get_r() {
            return true;
        }
        get r() {
            return this.get_r();
        }
        i(a, b) {
            return new WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter(a, b);
        }
        get_y() {
            return false;
        }
        constructor(a, b) {
            super(a, b);
        }
    }
    WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter.$t = /*@__PURE__*/ markType(WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter, 'WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter', CodeGeneratingComponentRendererCodeEmitter.$);
    return WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter;
})();
/**
 * @hidden
 */
export let CodeGeneratingComponentRenderer = /*@__PURE__*/ (() => {
    class CodeGeneratingComponentRenderer extends ComponentRenderer {
        loadCodeJson(a) {
            this.loadJson(a, (b) => ((() => {
                let $ret = new CodeGenerationContainer();
                $ret.id = b;
                return $ret;
            })()));
        }
        ds(a) {
            this.ex.c = a;
        }
        p() {
            return false;
        }
        static e5(a) {
            if (a == null) {
                return null;
            }
            if (a.length == 1) {
                return a.toLowerCase();
            }
            return a.substr(0, 1).toLowerCase() + a.substr(1);
        }
        static e7(a) {
            if (a == null) {
                return null;
            }
            if (a.length == 1) {
                return a.toLowerCase();
            }
            if (Base.equalsStatic(a.toUpperCase(), a)) {
                return a;
            }
            return a.substr(0, 1).toLowerCase() + a.substr(1);
        }
        static e6(a) {
            if (a == null) {
                return null;
            }
            if (a.length == 1) {
                return a.toUpperCase();
            }
            return a.substr(0, 1).toUpperCase() + a.substr(1);
        }
        static e8(a) {
            if (a == null) {
                return null;
            }
            let b = new List$1(String_$type, 0);
            let c = 0;
            for (let d = 0; d < a.length; d++) {
                let e = a.charAt(d);
                let f = e.toUpperCase();
                let g = e.toLowerCase();
                let h = (c == 0 && b.count > 0);
                let i = (c > 1);
                if (f == e) {
                    if (h) {
                        b.add('-');
                    }
                    c++;
                }
                else if (g == e) {
                    if (i) {
                        b.insert(b.count - 1, '-');
                    }
                    c = 0;
                }
                else {
                    c = 0;
                }
                b.add(g);
            }
            let j = new StringBuilder(0);
            for (let k = 0; k < b.count; k++) {
                j.h(b._inner[k]);
            }
            return j.toString();
        }
        at() {
            return true;
        }
        bz(a) {
            let b = a.g;
            return b == null ? { $type: Point_$type, x: NaN, y: NaN } : { $type: Point_$type, x: b.x, y: b.y };
        }
        b0(a) {
            let b = a.g;
            return b == null ? Rect.empty : new Rect(0, b.left, b.top, b.width, b.height);
        }
        b1(a) {
            let b = a.g;
            return b == null ? new Size(1, NaN, NaN) : new Size(1, b.width, b.height);
        }
        b3(a) {
            let b = a.g;
            if (typeof b === 'number') {
                return typeGetValue(b);
            }
            if (typeof b === 'number') {
                return b;
            }
            if (typeof b === 'string') {
                return parseFloat(b);
            }
            return NaN;
        }
        cu(a) {
            if (a == null) {
                return null;
            }
            return Brush.create(a);
        }
        cv(a) {
            if (a == null) {
                return null;
            }
            return Brush.create(a).color;
        }
        ag(a) {
            super.ag(a);
            this.e1 = a;
            return true;
        }
        an(a) {
            super.an(a);
            this.e3 = a;
            return true;
        }
        ah(a) {
            super.ah(a);
            this.ez = a;
            return true;
        }
        ai(a) {
            super.ai(a);
            this.e0 = a;
            return true;
        }
        aj(a) {
            super.aj(a);
            this.e2 = a;
            return true;
        }
        er(a, b, c) {
            switch (this.ey) {
                case 1: if (b.forceCodeBehind) {
                    return new AngularCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                else {
                    return new AngularTemplateCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                case 2: if (b.forceCodeBehind) {
                    return new ReactCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                else {
                    return new ReactMarkupCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                case 3: if (b.forceCodeBehind) {
                    return new WebComponentsCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                else {
                    return new WebComponentsMarkupCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                case 4: break;
                case 0: if (b.forceCodeBehind) {
                    return new WPFCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                else {
                    return new WPFXamlCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                case 5: if (b.forceCodeBehind) {
                    return new BlazorCodeBehindCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
                else {
                    return new BlazorRazorCodeGeneratingComponentRendererCodeEmitter(b, c);
                }
            }
            return null;
        }
        es(a, b) {
            switch (this.ey) {
                case 1: return new TSCodeGeneratingComponentRendererDataEmitter(a, b);
                case 2: return new TSCodeGeneratingComponentRendererDataEmitter(a, b);
                case 3: return new TSCodeGeneratingComponentRendererDataEmitter(a, b);
                case 4: break;
                case 0:
                case 5: return new DotNetCodeGeneratingComponentRendererDataEmitter(a, b);
            }
            return null;
        }
        et(a, b) {
            switch (this.ey) {
                case 1: return new TSCodeGeneratingComponentRendererHandlerEmitter(a, b, this.ey);
                case 2: return new TSCodeGeneratingComponentRendererHandlerEmitter(a, b, this.ey);
                case 3: return new TSCodeGeneratingComponentRendererHandlerEmitter(a, b, this.ey);
                case 4: break;
                case 0:
                case 5: return new DotNetCodeGeneratingComponentRendererHandlerEmitter(a, b);
            }
            return null;
        }
        ev(a, b) {
            switch (this.ey) {
                case 1: return new TSCodeGeneratingComponentRendererTemplateEmitter(a, b, this.ey);
                case 2: return new TSCodeGeneratingComponentRendererTemplateEmitter(a, b, this.ey);
                case 3: return new TSCodeGeneratingComponentRendererTemplateEmitter(a, b, this.ey);
                case 4: break;
                case 0:
                case 5: return new DotNetCodeGeneratingComponentRendererTemplateEmitter(a, b);
            }
            return null;
        }
        eu(a, b) {
            switch (this.ey) {
                case 1: return new TSCodeGeneratingComponentRendererModuleEmitter(this.ey, a, b);
                case 2: return new TSCodeGeneratingComponentRendererModuleEmitter(this.ey, a, b);
                case 3: return new TSCodeGeneratingComponentRendererModuleEmitter(this.ey, a, b);
                case 4: break;
                case 0: return new DotNetCodeGeneratingComponentRendererModuleEmitter(this.ey, a, b);
                case 5: return new BlazorCodeGeneratingComponentRendererModuleEmitter(this.ey, a, b);
            }
            return null;
        }
        markRefUsed(a) {
            for (let b of fromEnum(this.bc.keys)) {
                let c = b;
                c.markRefUsed(a);
            }
        }
        emitCode(a) {
            let b = new CodeGeneratingRendererResult();
            b.platform = this.ey;
            b.generationOptions = this.ex;
            for (let c of fromEnum(this.bc.keys)) {
                let d = c;
                let e = d.id;
                for (let f of fromEnum(d.b)) {
                    if (f.j("Name") != null) {
                        b.trackDescriptionField(e, f.j("Name"));
                    }
                    if (f.j("name") != null) {
                        b.trackDescriptionField(e, f.j("name"));
                    }
                }
            }
            for (let g of fromEnum(this.bc.keys)) {
                let h = g;
                h.d(this.context, b, this, this.ex);
            }
            if (this.e1 != null) {
                let i = this.eu(this.ex, b);
                i.n(this.e1, this.context);
            }
            if (this.e3 != null) {
                if (typeCast(JsonDictionaryObject.$, this.e3) !== null) {
                    let j = this.e3;
                    let m = j.e();
                    for (let l = 0; l < m.length; l++) {
                        let k = m[l];
                        b.defineString(k, j.item(k).value);
                    }
                }
            }
            {
                let n = this.es(this.ex, b);
                n.ac(this.e2);
            }
            let o = this.et(this.ex, b);
            let p = new List$1(String_$type, 0);
            let q = new List$1(String_$type, 0);
            if (this.ez != null) {
                if (typeCast(JsonDictionaryArray.$, this.ez) !== null) {
                    let r = this.ez;
                    if (r.items != null) {
                        let u = r.items;
                        for (let t = 0; t < u.length; t++) {
                            let s = u[t];
                            p.add(s.value.toString());
                        }
                    }
                }
                else if (typeCast(JsonDictionaryValue.$, this.ez) !== null) {
                    let v = this.ez;
                    p.add(v.value.toString());
                }
            }
            if (this.e0 != null) {
                if (typeCast(JsonDictionaryArray.$, this.e0) !== null) {
                    let w = this.e0;
                    if (w.items != null) {
                        let z = w.items;
                        for (let y = 0; y < z.length; y++) {
                            let x = z[y];
                            q.add(x.value.toString());
                        }
                    }
                }
                else if (typeCast(JsonDictionaryValue.$, this.e0) !== null) {
                    let aa = this.e0;
                    q.add(aa.value.toString());
                }
            }
            o.s(p.toArray(), q.toArray());
            let ab = this.ev(this.ex, b);
            ab.r();
            this.fd(b, a);
            return b;
        }
        fd(a, b) {
            b.execute(a);
        }
        bl() {
            return runOn(this, this.fa);
        }
        e9(a) {
            a = stringReplace(a, "-", "_");
            if (!this.ex.g) {
                switch (this.ey) {
                    case 1:
                    case 2:
                    case 3: return CodeGeneratingComponentRenderer.e5(a);
                    case 4:
                    case 0:
                    case 5: return CodeGeneratingComponentRenderer.e6(a);
                }
            }
            return a;
        }
        fa(a) {
            let b = a;
            if (stringStartsWith(a, "{")) {
                if (!stringStartsWith(a, "{[")) {
                    a = a.substr(1, a.length - 2);
                    a = "{" + this.e9(a) + "}";
                    b = a;
                }
            }
            else {
                b = this.e9(a);
            }
            return b;
        }
        constructor(a, b) {
            super();
            this.ex = null;
            this.ey = 0;
            this.e1 = null;
            this.e2 = null;
            this.e3 = null;
            this.e0 = null;
            this.ez = null;
            this.ex = b;
            this.ey = a;
            this.preserveKeyOrder = true;
            ComponentRenderer.platform = CodeGeneratingComponentRenderer.e4(a);
            this.adapter = new CodeGenerationRendererAdapter(a, this.adapter);
            if (b.reusedContext != null) {
                this.context = b.reusedContext;
            }
            else {
                this.context = new TypeDescriptionContext(this.adapter, ComponentRenderer.platform);
            }
            this.k = new CodeGenerationTransformer(a);
        }
        static e4(a) {
            switch (a) {
                case 1: return 3;
                case 2: return 2;
                case 3: return 4;
                case 4: return 8;
                case 0: return 0;
                case 5: return 12;
            }
            return 0;
        }
    }
    CodeGeneratingComponentRenderer.$t = /*@__PURE__*/ markType(CodeGeneratingComponentRenderer, 'CodeGeneratingComponentRenderer', ComponentRenderer.$);
    return CodeGeneratingComponentRenderer;
})();
/**
 * @hidden
 */
export let CodeGenerationContainer = /*@__PURE__*/ (() => {
    class CodeGenerationContainer extends Base {
        get id() {
            return this._id;
        }
        set id(a) {
            this._id = a;
        }
        constructor() {
            super();
            this._id = null;
            this.b = null;
            this.a = new HashSet$1(String_$type, 0);
            this.b = new List$1(CodeGenerationItemBuilder.$, 0);
        }
        d(a, b, c, d) {
            let e = new StringBuilder(0);
            let f = new StringBuilder(0);
            let g = new StringBuilder(0);
            let h = new CodeGeneratingImportManager();
            for (let i of fromEnum(this.a)) {
                if (!c.hasRef(i)) {
                    c.f(i);
                }
                b.af(((() => {
                    let $ret = new CodeGenerationItemBuilderPropertyInfo();
                    $ret.a = c.g(i);
                    return $ret;
                })()));
            }
            for (let j of fromEnum(this.b)) {
                let k = c.er(j, d, b);
                j.v(k, true);
                let l = k.toString();
                b.addContainerResult(this.id, l);
                k.be(e);
                k.bh(f);
                k.bd(g);
                k.bf(a, h);
            }
            b.addContainerResult("bindingFields", e.toString());
            b.addContainerResult("bindingInit", f.toString());
            b.addContainerResult("bindingCode", g.toString());
            b.addImportResult("bindingImports", h);
        }
        markRefUsed(a) {
            this.a.add_1(a);
        }
    }
    CodeGenerationContainer.$t = /*@__PURE__*/ markType(CodeGenerationContainer, 'CodeGenerationContainer');
    return CodeGenerationContainer;
})();
/**
 * @hidden
 */
export let CodeGenerationItemBuilder = /*@__PURE__*/ (() => {
    class CodeGenerationItemBuilder extends Base {
        p() {
            let a = this;
            let b = null;
            if (a.j("Name") != null) {
                b = a.j("Name");
            }
            if (a.j("name") != null) {
                b = a.j("name");
            }
            if (b != null) {
                return b;
            }
            return null;
        }
        constructor(a, b, c, d) {
            super();
            this.h = new List$1(String_$type, 0);
            this.g = new Dictionary$2(String_$type, CodeGenerationItemBuilderPropertyInfo.$, 0);
            this.c = null;
            this.b = 0;
            this.k = null;
            this.i = null;
            this.u = null;
            this.b = a;
            this.k = b;
            this.c = c;
            this.i = d;
        }
        a(a) {
            let b = this.c.getMetadata(this.k, a);
            return new CodeGenerationItemBuilder(this.b, this.s(b), this.c, this.i);
        }
        s(a) {
            return a.specificExternalType != null ? a.specificExternalType : a.specificType;
        }
        y(a, b, c, d) {
            if (!this.g.containsKey(a)) {
                this.h.add(a);
            }
            ;
            this.g.item(a, ((() => {
                let $ret = new CodeGenerationItemBuilderPropertyInfo();
                $ret.b = b;
                $ret.c = c;
                $ret.a = d;
                return $ret;
            })()));
        }
        w(a, b, c) {
            let d = null;
            if (!this.g.containsKey(a)) {
                this.h.add(a);
            }
            ;
            if (this.g.containsKey(a)) {
                let e = this.g.item(a);
                d = e.c;
            }
            this.g.item(a, ((() => {
                let $ret = new CodeGenerationItemBuilderPropertyInfo();
                $ret.b = b;
                $ret.c = d;
                $ret.a = c;
                return $ret;
            })()));
        }
        x(a, b) {
            if (this.g.containsKey(a)) {
                this.g.removeItem(a);
            }
        }
        j(a) {
            if (this.g.containsKey(a)) {
                return this.g.item(a).c;
            }
            return null;
        }
        v(a, b) {
            let c = this.n();
            if (b) {
                let d = this.q();
                a.a5 = d;
            }
            let e = false;
            let f = false;
            if (this.g.containsKey("Name")) {
                e = true;
                a.bp(this.g.item("Name").c);
            }
            else if (this.g.containsKey("name")) {
                e = true;
                a.bp(this.g.item("name").c);
            }
            else {
                if (b && this.u != null) {
                    e = true;
                    a.bp(this.u);
                }
                else if (this.f(this.c, a)) {
                    f = true;
                    a.bo(c);
                    if (a.y) {
                        this.h.add("name");
                        this.g.addItem("name", ((() => {
                            let $ret = new CodeGenerationItemBuilderPropertyInfo();
                            $ret.c = a.ap;
                            return $ret;
                        })()));
                    }
                    else {
                        this.h.add("Name");
                        this.g.addItem("Name", ((() => {
                            let $ret = new CodeGenerationItemBuilderPropertyInfo();
                            $ret.c = a.ap;
                            return $ret;
                        })()));
                    }
                }
            }
            a.bc(this.r(), this.k, this.c);
            let g = 0;
            for (let h of fromEnum(this.h)) {
                let i = this.g.item(h);
                if (i.b == null && h.toLowerCase() == "type") {
                    continue;
                }
                if (this.e(a, h, i)) {
                    a.bi(g, h, i, this.c);
                    g++;
                }
            }
            a.bq();
            let j = 0;
            for (let k of fromEnum(this.h)) {
                let l = this.g.item(k);
                if (l.b == null && k.toLowerCase() == "type") {
                    continue;
                }
                if (!this.e(a, k, l)) {
                    a.bj(j, k, l, this.c);
                    j++;
                }
            }
            a.bk(j);
            a.bl(c);
            if (e) {
                a.bn();
            }
            else {
                if (f) {
                    a.bm();
                }
            }
        }
        f(a, b) {
            if (b.n) {
                return true;
            }
            for (let c of fromEnum(this.h)) {
                let d = this.g.item(c);
                if (b.q(a, d)) {
                    return true;
                }
            }
            return false;
        }
        d(a) {
            let b = typeCast(Array_$type, a.c);
            for (let c of b) {
                if (typeCast(CodeGenerationItemBuilder.$, c) !== null) {
                    return false;
                }
            }
            return true;
        }
        e(a, b, c) {
            if (c.b == null) {
                if (b.toLowerCase() == "name") {
                    return true;
                }
            }
            switch (c.b.knownType) {
                case 10: return a.r ? this.d(c) : true;
                case 7: return true;
                case 4: return c.c == null || !(typeCast(LinearGradientBrush.$, c.c) !== null);
                case 6:
                    return true;
                    break;
                case 23: return true;
                case 9:
                    if (a.w(this.c, c)) {
                        return true;
                    }
                    return false;
                case 5: return true;
                case 16: return true;
                case 20: return true;
                case 22: return true;
                case 15: return false;
                case 3: return true;
                case 19: return true;
                case 8:
                    if (a.w(this.c, c)) {
                        return true;
                    }
                    if (c.b.specificExternalType == "string") {
                        return true;
                    }
                    if (c.a != null) {
                        return true;
                    }
                    return false;
                case 13: return false;
                case 18: return true;
                case 1: return true;
                case 11: return true;
                case 14: return true;
                case 12: return true;
                case 2: return true;
                case 21: return true;
                case 17: return true;
                case 0: return false;
            }
            return false;
        }
        r() {
            return this.n();
        }
        get t() {
            return this.k;
        }
        static o(a, b, c, d, e) {
            if (a == 0 || a == 4) {
                let f = c.getMetadata(b, "_qualifiedName");
                if (f != null) {
                    let g = f.specificExternalType;
                    let h = g.lastIndexOf('.');
                    if (h >= 0) {
                        return g.substr(h + 1);
                    }
                    else {
                        return g;
                    }
                }
                return b;
            }
            else if ((a == 5 || a == 2 || a == 1 || a == 3) && c.getMetadata(b, "__qualifiedNameTS") != null) {
                let i = c.getMetadata(b, "__qualifiedNameTS");
                if (c.getMetadata(b, "__skipSuffix") != null) {
                    return d + i.specificExternalType;
                }
                return d + i.specificExternalType + e;
            }
            else {
                if (c.getMetadata(b, "__skipSuffix") != null) {
                    return d + b;
                }
                return d + b + e;
            }
        }
        n() {
            return CodeGenerationItemBuilder.o(this.b, this.k, this.c, this.l, this.m);
        }
        q() {
            let a = this.c.getMetadata(this.k, "_qualifiedName");
            if (a != null) {
                let b = a.specificExternalType;
                let c = b.split('.');
                if (c.length > 2) {
                    return c[c.length - 2].toLowerCase();
                }
            }
            if (this.b == 1 || this.b == 3 || this.b == 2 || this.b == 5) {
                a = this.c.getMetadata(this.k, "__packageNameWC");
                if (a != null) {
                    let d = a.specificExternalType;
                    return d;
                }
            }
            if (this.k == "Legend" || this.k == "ItemLegend") {
                return "charts";
            }
            return "core";
        }
        get_l() {
            switch (this.b) {
                case 1: return "Igx";
                case 2: return "Igr";
                case 3: return "Igc";
                case 4: return "Ultra";
                case 0: return "Xam";
                case 5: return "Igb";
            }
            return "";
        }
        get l() {
            return this.get_l();
        }
        get_m() {
            switch (this.b) {
                case 1: return "Component";
                case 2: return "";
                case 3: return "Component";
                case 4: return "";
                case 0: return "";
            }
            return "";
        }
        get m() {
            return this.get_m();
        }
    }
    CodeGenerationItemBuilder.$t = /*@__PURE__*/ markType(CodeGenerationItemBuilder, 'CodeGenerationItemBuilder');
    return CodeGenerationItemBuilder;
})();
/**
 * @hidden
 */
export let CodeGenerationRendererAdapter = /*@__PURE__*/ (() => {
    class CodeGenerationRendererAdapter extends Base {
        constructor(a, b) {
            super();
            this.a = 0;
            this.b = null;
            this.b = b;
            this.a = a;
        }
        createColorCollection(a) {
            let b = new List$1(Base.$, 0);
            for (let c = 0; c < a.length; c++) {
                b.add1(a[c]);
            }
            return b;
        }
        createBrushCollection(a) {
            let b = new List$1(Base.$, 0);
            for (let c = 0; c < a.length; c++) {
                b.add1(a[c]);
            }
            return b;
        }
        createDoubleCollection(a) {
            let b = new List$1(Base.$, 0);
            for (let c = 0; c < a.length; c++) {
                b.add1(a[c]);
            }
            return b.toArray();
        }
        createObject(a, b, c, d) {
            let e = new CodeGenerationItemBuilder(this.a, a, c, b);
            return e;
        }
        coerceToEnum(a, b, c) {
            return c;
        }
        onUIThread(a, b) {
            b();
        }
        setOrUpdateCollectionOnTarget(a, b, c, d, e, f) {
            let g = this.getPropertyValue(e, b);
            if (g == null) {
                let h = new List$1(Base.$, 0);
                let i = f;
                for (let j = 0; j < i.count; j++) {
                    h.add1(i.item(j));
                }
                this.setPropertyValue(e, b, c, h, null, null);
            }
            else {
                let k = f;
                let l = g;
                l.clear();
                for (let m = 0; m < k.count; m++) {
                    l.add(k.item(m));
                }
            }
        }
        c(a) {
            if (a == null) {
                return false;
            }
            return false;
        }
        onPendingRef(a, b, c, d) {
            let e = a;
            e.w(b, c, d);
        }
        setPropertyValue(a, b, c, d, e, f) {
            if (!(typeCast(CodeGenerationItemBuilder.$, a) !== null)) {
                this.b.setPropertyValue(a, b, c, d, e, f);
                return;
            }
            while (stringContains(b, ".")) {
                let g = b.split('.');
                let h = this.getPropertyValue(a, g[0]);
                if (h != null && !this.c(h)) {
                    b = "";
                    for (let i = 1; i < g.length; i++) {
                        if (i > 1) {
                            b += ".";
                        }
                        b += g[i];
                    }
                    a = h;
                }
                else {
                    let j = a;
                    let k = j.a(g[0]);
                    h = k;
                    this.setPropertyValue(a, g[0], c, h, null, f);
                    b = "";
                    for (let l = 1; l < g.length; l++) {
                        if (l > 1) {
                            b += ".";
                        }
                        b += g[l];
                    }
                    a = h;
                }
            }
            let m = a;
            m.y(b, c, d, f);
        }
        getPropertyValue(a, b) {
            while (stringContains(b, ".")) {
                let c = b.split('.');
                let d = this.getPropertyValue(a, c[0]);
                if (d != null) {
                    b = "";
                    for (let e = 1; e < c.length; e++) {
                        if (e > 1) {
                            b += ".";
                        }
                        b += c[e];
                    }
                    a = d;
                }
                else {
                    let f = a;
                    let g = f.a(c[0]);
                    d = g;
                    for (let h = 1; h < c.length; h++) {
                        if (h > 1) {
                            b += ".";
                        }
                        b += c[h];
                    }
                    a = d;
                }
            }
            let i = a;
            return i.j(b);
        }
        getRootObject(a) {
            let b = a;
            let c = b.b.count > 0 ? b.b._inner[0] : null;
            return c;
        }
        clearContainer(a, b, c) {
            let d = a;
            let e = d.b.count > 0 ? d.b._inner[0] : null;
            d.b.clear();
            c(false);
        }
        forPropertyValueItem(a, b, c) {
            let d = this.getPropertyValue(a, b);
            if (typeCast(IList_$type, d) !== null) {
                let e = d;
                for (let f = 0; f < e.count; f++) {
                    c(e.item(f));
                }
            }
            else if (typeCast(Array_$type, d) !== null) {
                let g = d;
                for (let h = 0; h < g.length; h++) {
                    c(g[h]);
                }
            }
        }
        clearCollection(a, b, c) {
            let d = this.getPropertyValue(a, b);
            if (typeCast(IList_$type, d) !== null) {
                d.clear();
            }
            else if (typeCast(Array_$type, d) !== null) {
                this.setPropertyValue(a, b, c, null, d, null);
            }
        }
        addItemToCollection(a, b, c, d, e) {
            let f = this.getPropertyValue(c, a);
            if (typeCast(IList_$type, f) !== null) {
                f.insert(d, e);
            }
            else if (typeCast(Array_$type, f) !== null) {
                let g = f;
                let h = new List$1(Base.$, 0);
                for (let i = 0; i < g.count; i++) {
                    h.add1(g.item(i));
                }
                h.insert1(d, e);
                let j = [g.count + 1];
                for (let k = 0; k < j.length; k++) {
                    j[k] = h._inner[k];
                }
                this.setPropertyValue(c, a, b, j, f, null);
            }
        }
        resetPropertyOnTarget(a, b, c, d) {
            while (stringContains(b, ".")) {
                let e = b.split('.');
                let f = this.getPropertyValue(d, e[0]);
                if (f != null && !this.c(f)) {
                    b = "";
                    for (let g = 1; g < e.length; g++) {
                        if (g > 1) {
                            b += ".";
                        }
                        b += e[g];
                    }
                    d = f;
                }
                else {
                    let h = d;
                    let i = h.a(e[0]);
                    f = i;
                    this.setPropertyValue(d, e[0], c, f, null, null);
                    b = "";
                    for (let j = 1; j < e.length; j++) {
                        if (j > 1) {
                            b += ".";
                        }
                        b += e[j];
                    }
                    d = f;
                }
            }
            let k = d;
            k.x(b, c);
        }
        replaceItemInCollection(a, b, c, d, e) {
            let f = this.getPropertyValue(c, a);
            if (typeCast(IList_$type, f) !== null) {
                f.item(d, e);
            }
            else if (typeCast(Array_$type, f) !== null) {
                let g = f;
                let h = new List$1(Base.$, 0);
                for (let i = 0; i < g.count; i++) {
                    h.add1(g.item(i));
                }
                h._inner[d] = e;
                let j = [g.count + 1];
                for (let k = 0; k < j.length; k++) {
                    j[k] = h._inner[k];
                }
                this.setPropertyValue(c, a, b, j, f, null);
            }
        }
        replaceRootItem(a, b, c, d) {
            let e = this.createObject(b, a, c, null);
            let f = a;
            let g = f.b.count > 0 ? f.b._inner[0] : null;
            f.b.clear();
            {
                f.b.add(e);
            }
            d(false);
        }
        removeItemFromCollection(a, b, c, d) {
            let e = this.getPropertyValue(c, a);
            if (typeCast(IList_$type, e) !== null) {
                e.removeAt(d);
            }
            else if (typeCast(Array_$type, e) !== null) {
                let f = e;
                let g = new List$1(Base.$, 0);
                for (let h = 0; h < f.count; h++) {
                    g.add1(f.item(h));
                }
                g.removeAt(d);
                let i = [f.count - 1];
                for (let j = 0; j < i.length; j++) {
                    i[j] = g._inner[j];
                }
                this.setPropertyValue(c, a, b, i, e, null);
            }
        }
        removeRootItem(a, b, c) {
            let d = a;
            let e = d.b.count > 0 ? d.b._inner[0] : null;
            d.b.clear();
            c(false);
        }
        flushChanges(a) {
        }
        executeMethod(a, b, c, d, e) {
        }
        serializeBrush(a) {
            return null;
        }
        serializeColor(a) {
            return null;
        }
        serializeBrushCollection(a) {
            return null;
        }
        serializePoint(a) {
            return null;
        }
        serializeSize(a) {
            return null;
        }
        serializeRect(a) {
            return null;
        }
        serializeColorCollection(a) {
            return null;
        }
        serializeTimespan(a) {
            return null;
        }
        serializeDoubleCollection(a) {
            return null;
        }
    }
    CodeGenerationRendererAdapter.$t = /*@__PURE__*/ markType(CodeGenerationRendererAdapter, 'CodeGenerationRendererAdapter', Base.$, [IComponentRendererAdapter_$type]);
    return CodeGenerationRendererAdapter;
})();
