// igniteui-react-core/esm2015/lib/CodeGenerationTargetPlatforms.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
export var CodeGenerationTargetPlatforms = /*@__PURE__*/ (function (CodeGenerationTargetPlatforms) {
    CodeGenerationTargetPlatforms[CodeGenerationTargetPlatforms["WPF"] = 0] = "WPF";
    CodeGenerationTargetPlatforms[CodeGenerationTargetPlatforms["Angular"] = 1] = "Angular";
    CodeGenerationTargetPlatforms[CodeGenerationTargetPlatforms["React"] = 2] = "React";
    CodeGenerationTargetPlatforms[CodeGenerationTargetPlatforms["WebComponents"] = 3] = "WebComponents";
    CodeGenerationTargetPlatforms[CodeGenerationTargetPlatforms["WindowsForms"] = 4] = "WindowsForms";
    CodeGenerationTargetPlatforms[CodeGenerationTargetPlatforms["Blazor"] = 5] = "Blazor";
    return CodeGenerationTargetPlatforms;
})({});
/**
 * @hidden
 */
export let CodeGenerationTargetPlatforms_$type = /*@__PURE__*/ markEnum('CodeGenerationTargetPlatforms', 'WPF,0|Angular,1|React,2|WebComponents,3|WindowsForms,4|Blazor,5');
