// igniteui-react-core/esm2015/lib/CodeGeneratingComponentRendererDataEmitterPropertyType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let CodeGeneratingComponentRendererDataEmitterPropertyType = /*@__PURE__*/ (() => {
    class CodeGeneratingComponentRendererDataEmitterPropertyType extends Base {
        constructor() {
            super(...arguments);
            this.c = 0;
            this.f = false;
            this.a = null;
            this.b = null;
            this.d = false;
        }
        e() {
            if (this.c == 3) {
                return false;
            }
            if (!this.f && !this.d) {
                return true;
            }
            if (this.a == null) {
                return false;
            }
            for (let a = 0; a < this.a.length; a++) {
                let b = this.b[a];
                if (b.f) {
                    if (!b.e()) {
                        return false;
                    }
                }
            }
            return true;
        }
    }
    CodeGeneratingComponentRendererDataEmitterPropertyType.$t = /*@__PURE__*/ markType(CodeGeneratingComponentRendererDataEmitterPropertyType, 'CodeGeneratingComponentRendererDataEmitterPropertyType');
    return CodeGeneratingComponentRendererDataEmitterPropertyType;
})();
