// igniteui-react-core/esm2015/lib/CodeGenerationLibraryItemContentLocation.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
export var CodeGenerationLibraryItemContentLocation = /*@__PURE__*/ (function (CodeGenerationLibraryItemContentLocation) {
    CodeGenerationLibraryItemContentLocation[CodeGenerationLibraryItemContentLocation["Code"] = 0] = "Code";
    CodeGenerationLibraryItemContentLocation[CodeGenerationLibraryItemContentLocation["JsonFile"] = 1] = "JsonFile";
    CodeGenerationLibraryItemContentLocation[CodeGenerationLibraryItemContentLocation["CDN"] = 2] = "CDN";
    CodeGenerationLibraryItemContentLocation[CodeGenerationLibraryItemContentLocation["CDNSkipHydrate"] = 3] = "CDNSkipHydrate";
    return CodeGenerationLibraryItemContentLocation;
})({});
/**
 * @hidden
 */
export let CodeGenerationLibraryItemContentLocation_$type = /*@__PURE__*/ markEnum('CodeGenerationLibraryItemContentLocation', 'Code,0|JsonFile,1|CDN,2|CDNSkipHydrate,3');
