// igniteui-react-core/esm2015/lib/CodeGenerationLibraryItemPlatform.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
export var CodeGenerationLibraryItemPlatform = /*@__PURE__*/ (function (CodeGenerationLibraryItemPlatform) {
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["Angular"] = 0] = "Angular";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["React"] = 1] = "React";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["WebComponents"] = 2] = "WebComponents";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["JQuery"] = 3] = "JQuery";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["WPF"] = 4] = "WPF";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["WindowsForms"] = 5] = "WindowsForms";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["Web"] = 6] = "Web";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["AllWeb"] = 7] = "AllWeb";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["Desktop"] = 8] = "Desktop";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["Win"] = 9] = "Win";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["XPlat"] = 10] = "XPlat";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["All"] = 11] = "All";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["Unknown"] = 12] = "Unknown";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["DotNet"] = 13] = "DotNet";
    CodeGenerationLibraryItemPlatform[CodeGenerationLibraryItemPlatform["Blazor"] = 14] = "Blazor";
    return CodeGenerationLibraryItemPlatform;
})({});
/**
 * @hidden
 */
export let CodeGenerationLibraryItemPlatform_$type = /*@__PURE__*/ markEnum('CodeGenerationLibraryItemPlatform', 'Angular,0|React,1|WebComponents,2|JQuery,3|WPF,4|WindowsForms,5|Web,6|AllWeb,7|Desktop,8|Win,9|XPlat,10|All,11|Unknown,12|DotNet,13|Blazor,14');
