// igniteui-react-core/esm2015/lib/CollisionGeometryType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
export var CollisionGeometryType = /*@__PURE__*/ (function (CollisionGeometryType) {
    CollisionGeometryType[CollisionGeometryType["Box"] = 0] = "Box";
    CollisionGeometryType[CollisionGeometryType["Boxes"] = 1] = "Boxes";
    CollisionGeometryType[CollisionGeometryType["PieSlice"] = 2] = "PieSlice";
    CollisionGeometryType[CollisionGeometryType["Circle"] = 3] = "Circle";
    return CollisionGeometryType;
})({});
/**
 * @hidden
 */
export let CollisionGeometryType_$type = /*@__PURE__*/ markEnum('CollisionGeometryType', 'Box,0|Boxes,1|PieSlice,2|Circle,3');
