// igniteui-react-core/esm2015/lib/ColumnWidthDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnWidthDescription = /*@__PURE__*/ (() => {
    class ColumnWidthDescription extends Description {
        get_type() {
            return "ColumnWidth";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = false;
            this.l = 0;
            this.k = 0;
        }
        get isStarSized() {
            return this.i;
        }
        set isStarSized(a) {
            this.i = a;
            this.g("IsStarSized");
        }
        get value() {
            return this.l;
        }
        set value(a) {
            this.l = a;
            this.g("Value");
        }
        get minimumWidth() {
            return this.k;
        }
        set minimumWidth(a) {
            this.k = a;
            this.g("MinimumWidth");
        }
    }
    ColumnWidthDescription.$t = /*@__PURE__*/ markType(ColumnWidthDescription, 'ColumnWidthDescription', Description.$);
    ColumnWidthDescription.__marshalByValue = true;
    ColumnWidthDescription.__marshalByValueAlias = "ColumnWidth";
    return ColumnWidthDescription;
})();
