// igniteui-react-core/esm2015/lib/ColumnPinnedChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnPinnedChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class ColumnPinnedChangedEventArgsDescription extends Description {
        get_type() {
            return "ColumnPinnedChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = null;
        }
        get column() {
            return this.h;
        }
        set column(a) {
            this.h = a;
            this.g("Column");
        }
        get pinned() {
            return this.j;
        }
        set pinned(a) {
            this.j = a;
            this.g("Pinned");
        }
    }
    ColumnPinnedChangedEventArgsDescription.$t = /*@__PURE__*/ markType(ColumnPinnedChangedEventArgsDescription, 'ColumnPinnedChangedEventArgsDescription', Description.$);
    return ColumnPinnedChangedEventArgsDescription;
})();
