// igniteui-react-core/esm2015/lib/ColumnResizingSeparatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnResizingSeparatorDescription = /*@__PURE__*/ (() => {
    class ColumnResizingSeparatorDescription extends DefinitionBaseDescription {
        get_type() {
            return "ColumnResizingSeparator";
        }
        constructor() {
            super();
            this.co = 0;
            this.cp = 0;
        }
        get actualOpacity() {
            return this.co;
        }
        set actualOpacity(a) {
            this.co = a;
            this.g("ActualOpacity");
        }
        get opacity() {
            return this.cp;
        }
        set opacity(a) {
            this.cp = a;
            this.g("Opacity");
        }
    }
    ColumnResizingSeparatorDescription.$t = /*@__PURE__*/ markType(ColumnResizingSeparatorDescription, 'ColumnResizingSeparatorDescription', DefinitionBaseDescription.$);
    return ColumnResizingSeparatorDescription;
})();
