// igniteui-react-core/esm2015/lib/ColumnGroupingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnGroupingDescription = /*@__PURE__*/ (() => {
    class ColumnGroupingDescription extends Description {
        get_type() {
            return "ColumnGrouping";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.s = null;
            this.k = null;
            this.o = null;
            this.u = null;
            this.v = null;
            this.l = null;
            this.h = 0;
            this.m = null;
            this.n = null;
            this.t = null;
            this.p = null;
            this.q = null;
            this.r = null;
            this.j = null;
        }
        get targetGridRef() {
            return this.s;
        }
        set targetGridRef(a) {
            this.s = a;
            this.g("TargetGridRef");
        }
        get baseTheme() {
            return this.k;
        }
        set baseTheme(a) {
            this.k = a;
            this.g("BaseTheme");
        }
        get density() {
            return this.o;
        }
        set density(a) {
            this.o = a;
            this.g("Density");
        }
        get title() {
            return this.u;
        }
        set title(a) {
            this.u = a;
            this.g("Title");
        }
        get titleColor() {
            return this.v;
        }
        set titleColor(a) {
            this.v = a;
            this.g("TitleColor");
        }
        get columnGroupTitleFontFamily() {
            return this.l;
        }
        set columnGroupTitleFontFamily(a) {
            this.l = a;
            this.g("ColumnGroupTitleFontFamily");
        }
        get columnGroupTitleFontSize() {
            return this.h;
        }
        set columnGroupTitleFontSize(a) {
            this.h = a;
            this.g("ColumnGroupTitleFontSize");
        }
        get columnGroupTitleFontStyle() {
            return this.m;
        }
        set columnGroupTitleFontStyle(a) {
            this.m = a;
            this.g("ColumnGroupTitleFontStyle");
        }
        get columnGroupTitleFontWeight() {
            return this.n;
        }
        set columnGroupTitleFontWeight(a) {
            this.n = a;
            this.g("ColumnGroupTitleFontWeight");
        }
        get textColor() {
            return this.t;
        }
        set textColor(a) {
            this.t = a;
            this.g("TextColor");
        }
        get iconColor() {
            return this.p;
        }
        set iconColor(a) {
            this.p = a;
            this.g("IconColor");
        }
        get itemBackgroundColor() {
            return this.q;
        }
        set itemBackgroundColor(a) {
            this.q = a;
            this.g("ItemBackgroundColor");
        }
        get itemHoverBackgroundColor() {
            return this.r;
        }
        set itemHoverBackgroundColor(a) {
            this.r = a;
            this.g("ItemHoverBackgroundColor");
        }
        get backgroundColor() {
            return this.j;
        }
        set backgroundColor(a) {
            this.j = a;
            this.g("BackgroundColor");
        }
    }
    ColumnGroupingDescription.$t = /*@__PURE__*/ markType(ColumnGroupingDescription, 'ColumnGroupingDescription', Description.$);
    return ColumnGroupingDescription;
})();
