// igniteui-react-core/esm2015/lib/ColumnFilterConditionGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ColumnFilterConditionDescription } from "./ColumnFilterConditionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnFilterConditionGroupDescription = /*@__PURE__*/ (() => {
    class ColumnFilterConditionGroupDescription extends ColumnFilterConditionDescription {
        get_type() {
            return "ColumnFilterConditionGroup";
        }
        constructor() {
            super();
            this.k = false;
            this.i = null;
        }
        get usesOrOperator() {
            return this.k;
        }
        set usesOrOperator(a) {
            this.k = a;
            this.g("UsesOrOperator");
        }
        get item() {
            return this.i;
        }
        set item(a) {
            this.i = a;
            this.g("Item");
        }
    }
    ColumnFilterConditionGroupDescription.$t = /*@__PURE__*/ markType(ColumnFilterConditionGroupDescription, 'ColumnFilterConditionGroupDescription', ColumnFilterConditionDescription.$);
    return ColumnFilterConditionGroupDescription;
})();
