// igniteui-react-core/esm2015/lib/ColumnPropertySetterDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnPropertySetterDescription = /*@__PURE__*/ (() => {
    class ColumnPropertySetterDescription extends Description {
        get_type() {
            return "ColumnPropertySetter";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.i = null;
            this.j = null;
        }
        get columnName() {
            return this.h;
        }
        set columnName(a) {
            this.h = a;
            this.g("ColumnName");
        }
        get propertyName() {
            return this.i;
        }
        set propertyName(a) {
            this.i = a;
            this.g("PropertyName");
        }
        get valueRef() {
            return this.j;
        }
        set valueRef(a) {
            this.j = a;
            this.g("ValueRef");
        }
    }
    ColumnPropertySetterDescription.$t = /*@__PURE__*/ markType(ColumnPropertySetterDescription, 'ColumnPropertySetterDescription', Description.$);
    return ColumnPropertySetterDescription;
})();
