// igniteui-react-core/esm2015/lib/ColumnChooserDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnChooserDescription = /*@__PURE__*/ (() => {
    class ColumnChooserDescription extends Description {
        get_type() {
            return "ColumnChooser";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.o = null;
            this.q = null;
            this.h = 0;
            this.r = null;
            this.s = null;
            this.p = null;
            this.m = null;
            this.k = null;
            this.l = null;
            this.j = null;
        }
        get targetGridRef() {
            return this.n;
        }
        set targetGridRef(a) {
            this.n = a;
            this.g("TargetGridRef");
        }
        get title() {
            return this.o;
        }
        set title(a) {
            this.o = a;
            this.g("Title");
        }
        get titleFontFamily() {
            return this.q;
        }
        set titleFontFamily(a) {
            this.q = a;
            this.g("TitleFontFamily");
        }
        get titleFontSize() {
            return this.h;
        }
        set titleFontSize(a) {
            this.h = a;
            this.g("TitleFontSize");
        }
        get titleFontStyle() {
            return this.r;
        }
        set titleFontStyle(a) {
            this.r = a;
            this.g("TitleFontStyle");
        }
        get titleFontWeight() {
            return this.s;
        }
        set titleFontWeight(a) {
            this.s = a;
            this.g("TitleFontWeight");
        }
        get titleColor() {
            return this.p;
        }
        set titleColor(a) {
            this.p = a;
            this.g("TitleColor");
        }
        get filterPlaceholderText() {
            return this.m;
        }
        set filterPlaceholderText(a) {
            this.m = a;
            this.g("FilterPlaceholderText");
        }
        get baseTheme() {
            return this.k;
        }
        set baseTheme(a) {
            this.k = a;
            this.g("BaseTheme");
        }
        get density() {
            return this.l;
        }
        set density(a) {
            this.l = a;
            this.g("Density");
        }
        get backgroundColor() {
            return this.j;
        }
        set backgroundColor(a) {
            this.j = a;
            this.g("BackgroundColor");
        }
    }
    ColumnChooserDescription.$t = /*@__PURE__*/ markType(ColumnChooserDescription, 'ColumnChooserDescription', Description.$);
    return ColumnChooserDescription;
})();
