// igniteui-react-core/esm2015/lib/ColumnFragmentDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FragmentBaseDescription } from "./FragmentBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnFragmentDescription = /*@__PURE__*/ (() => {
    class ColumnFragmentDescription extends FragmentBaseDescription {
        get_type() {
            return "ColumnFragment";
        }
        constructor() {
            super();
            this.gz = 0;
            this.g0 = 0;
        }
        get radiusX() {
            return this.gz;
        }
        set radiusX(a) {
            this.gz = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.g0;
        }
        set radiusY(a) {
            this.g0 = a;
            this.g("RadiusY");
        }
    }
    ColumnFragmentDescription.$t = /*@__PURE__*/ markType(ColumnFragmentDescription, 'ColumnFragmentDescription', FragmentBaseDescription.$);
    return ColumnFragmentDescription;
})();
