// igniteui-react-core/esm2015/lib/ColumnHiddenChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnHiddenChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class ColumnHiddenChangedEventArgsDescription extends Description {
        get_type() {
            return "ColumnHiddenChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = false;
        }
        get column() {
            return this.h;
        }
        set column(a) {
            this.h = a;
            this.g("Column");
        }
        get isHidden() {
            return this.j;
        }
        set isHidden(a) {
            this.j = a;
            this.g("IsHidden");
        }
    }
    ColumnHiddenChangedEventArgsDescription.$t = /*@__PURE__*/ markType(ColumnHiddenChangedEventArgsDescription, 'ColumnHiddenChangedEventArgsDescription', Description.$);
    return ColumnHiddenChangedEventArgsDescription;
})();
