// igniteui-react-core/esm2015/lib/ColumnGroupDescriptionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ColumnSortDescriptionDescription } from "./ColumnSortDescriptionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnGroupDescriptionDescription = /*@__PURE__*/ (() => {
    class ColumnGroupDescriptionDescription extends ColumnSortDescriptionDescription {
        get_type() {
            return "ColumnGroupDescription";
        }
        constructor() {
            super();
            this.x = null;
            this.p = null;
            this.u = null;
            this.o = null;
            this.v = null;
            this.w = null;
        }
        get valueFormat() {
            return this.x;
        }
        set valueFormat(a) {
            this.x = a;
            this.g("ValueFormat");
        }
        get valueFormatSpecifiers() {
            return this.p;
        }
        set valueFormatSpecifiers(a) {
            this.p = a;
            this.g("ValueFormatSpecifiers");
        }
        get displayFormat() {
            return this.u;
        }
        set displayFormat(a) {
            this.u = a;
            this.g("DisplayFormat");
        }
        get displayFormatSpecifiers() {
            return this.o;
        }
        set displayFormatSpecifiers(a) {
            this.o = a;
            this.g("DisplayFormatSpecifiers");
        }
        get displayName() {
            return this.v;
        }
        set displayName(a) {
            this.v = a;
            this.g("DisplayName");
        }
        get formatTextRef() {
            return this.w;
        }
        set formatTextRef(a) {
            this.w = a;
            this.g("FormatTextRef");
        }
    }
    ColumnGroupDescriptionDescription.$t = /*@__PURE__*/ markType(ColumnGroupDescriptionDescription, 'ColumnGroupDescriptionDescription', ColumnSortDescriptionDescription.$);
    ColumnGroupDescriptionDescription.__marshalByValue1 = true;
    ColumnGroupDescriptionDescription.__marshalByValueAlias1 = "ColumnGroupDescription";
    return ColumnGroupDescriptionDescription;
})();
