// igniteui-react-core/esm2015/lib/ComboBoxColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DataGridColumnDescription } from "./DataGridColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ComboBoxColumnDescription = /*@__PURE__*/ (() => {
    class ComboBoxColumnDescription extends DataGridColumnDescription {
        get_type() {
            return "ComboBoxColumn";
        }
        constructor() {
            super();
            this.f5 = null;
            this.f1 = null;
            this.f6 = null;
            this.f7 = null;
        }
        get dataSourceRef() {
            return this.f5;
        }
        set dataSourceRef(a) {
            this.f5 = a;
            this.g("DataSourceRef");
        }
        get dataSource() {
            return this.f1;
        }
        set dataSource(a) {
            this.f1 = a;
            this.g("DataSource");
        }
        get textField() {
            return this.f6;
        }
        set textField(a) {
            this.f6 = a;
            this.g("TextField");
        }
        get valueField() {
            return this.f7;
        }
        set valueField(a) {
            this.f7 = a;
            this.g("ValueField");
        }
    }
    ComboBoxColumnDescription.$t = /*@__PURE__*/ markType(ComboBoxColumnDescription, 'ComboBoxColumnDescription', DataGridColumnDescription.$);
    ComboBoxColumnDescription.__canMarshalByValue1 = true;
    ComboBoxColumnDescription.__marshalByValueAlias1 = "ComboBoxColumn";
    return ComboBoxColumnDescription;
})();
