// igniteui-react-core/esm2015/lib/CornerRadius.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, typeCast, markType } from "./type";
/**
 * @hidden
 */
export let CornerRadius = /*@__PURE__*/ (() => {
    class CornerRadius extends Base {
        constructor(a, ..._rest) {
            super();
            this.c = 0;
            this.b = 0;
            this.d = 0;
            this.e = 0;
            a = (a == void 0) ? 0 : a;
            switch (a) {
                case 0:
                    {
                        let c = _rest[0];
                        this.b = this.c = this.d = this.e = c;
                    }
                    break;
                case 1:
                    {
                        let c = _rest[0];
                        let d = _rest[1];
                        let e = _rest[2];
                        let f = _rest[3];
                        this.d = c;
                        this.e = d;
                        this.c = e;
                        this.b = f;
                    }
                    break;
                case 2:
                    {
                        this.b = NaN;
                        this.c = NaN;
                        this.d = NaN;
                        this.e = NaN;
                    }
                    break;
            }
        }
        equals(a) {
            if (a == null || !(typeCast(CornerRadius.$, a) !== null)) {
                return super.equals(a);
            }
            let b = a;
            return b.b == this.b && b.d == this.d && b.e == this.e && b.c == this.c;
        }
        getHashCode() {
            return (this.d) ^ (this.e) ^ (this.b) ^ (this.c);
        }
    }
    CornerRadius.$t = /*@__PURE__*/ markType(CornerRadius, 'CornerRadius');
    return CornerRadius;
})();
