// igniteui-react-core/esm2015/lib/CustomPaletteColorScaleDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ColorScaleDescription } from "./ColorScaleDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CustomPaletteColorScaleDescription = /*@__PURE__*/ (() => {
    class CustomPaletteColorScaleDescription extends ColorScaleDescription {
        get_type() {
            return "CustomPaletteColorScale";
        }
        constructor() {
            super();
            this.n = 0;
            this.m = 0;
            this.k = null;
            this.q = null;
        }
        get minimumValue() {
            return this.n;
        }
        set minimumValue(a) {
            this.n = a;
            this.g("MinimumValue");
        }
        get maximumValue() {
            return this.m;
        }
        set maximumValue(a) {
            this.m = a;
            this.g("MaximumValue");
        }
        get palette() {
            return this.k;
        }
        set palette(a) {
            this.k = a;
            this.g("Palette");
        }
        get interpolationMode() {
            return this.q;
        }
        set interpolationMode(a) {
            this.q = a;
            this.g("InterpolationMode");
        }
    }
    CustomPaletteColorScaleDescription.$t = /*@__PURE__*/ markType(CustomPaletteColorScaleDescription, 'CustomPaletteColorScaleDescription', ColorScaleDescription.$);
    return CustomPaletteColorScaleDescription;
})();
