// igniteui-react-core/esm2015/lib/DashboardTileCustomizationDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DashboardTileCustomizationDescription = /*@__PURE__*/ (() => {
    class DashboardTileCustomizationDescription extends Description {
        get_type() {
            return "DashboardTileCustomization";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.v = null;
            this.w = null;
            this.x = null;
            this.h = false;
            this.j = false;
            this.i = false;
            this.k = false;
            this.p = 0;
            this.r = 0;
            this.q = 0;
        }
        get matchType() {
            return this.v;
        }
        set matchType(a) {
            this.v = a;
            this.g("MatchType");
        }
        get propertyName() {
            return this.w;
        }
        set propertyName(a) {
            this.w = a;
            this.g("PropertyName");
        }
        get propertyValueRef() {
            return this.x;
        }
        set propertyValueRef(a) {
            this.x = a;
            this.g("PropertyValueRef");
        }
        get isCollectionInsertionAtEnd() {
            return this.h;
        }
        set isCollectionInsertionAtEnd(a) {
            this.h = a;
            this.g("IsCollectionInsertionAtEnd");
        }
        get isCollectionInsertionAtStart() {
            return this.j;
        }
        set isCollectionInsertionAtStart(a) {
            this.j = a;
            this.g("IsCollectionInsertionAtStart");
        }
        get isCollectionInsertionAtIndex() {
            return this.i;
        }
        set isCollectionInsertionAtIndex(a) {
            this.i = a;
            this.g("IsCollectionInsertionAtIndex");
        }
        get isCollectionRemovaltIndex() {
            return this.k;
        }
        set isCollectionRemovaltIndex(a) {
            this.k = a;
            this.g("IsCollectionRemovaltIndex");
        }
        get collectionIndex() {
            return this.p;
        }
        set collectionIndex(a) {
            this.p = a;
            this.g("CollectionIndex");
        }
        get order() {
            return this.r;
        }
        set order(a) {
            this.r = a;
            this.g("Order");
        }
        get matchParentIndex() {
            return this.q;
        }
        set matchParentIndex(a) {
            this.q = a;
            this.g("MatchParentIndex");
        }
    }
    DashboardTileCustomizationDescription.$t = /*@__PURE__*/ markType(DashboardTileCustomizationDescription, 'DashboardTileCustomizationDescription', Description.$);
    return DashboardTileCustomizationDescription;
})();
