// igniteui-react-core/esm2015/lib/DashboardTileFilterStringErrorsParsingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DashboardTileFilterStringErrorsParsingEventArgsDescription = /*@__PURE__*/ (() => {
    class DashboardTileFilterStringErrorsParsingEventArgsDescription extends Description {
        get_type() {
            return "DashboardTileFilterStringErrorsParsingEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
        }
        get propertyName() {
            return this.i;
        }
        set propertyName(a) {
            this.i = a;
            this.g("PropertyName");
        }
        get errors() {
            return this.h;
        }
        set errors(a) {
            this.h = a;
            this.g("Errors");
        }
    }
    DashboardTileFilterStringErrorsParsingEventArgsDescription.$t = /*@__PURE__*/ markType(DashboardTileFilterStringErrorsParsingEventArgsDescription, 'DashboardTileFilterStringErrorsParsingEventArgsDescription', Description.$);
    return DashboardTileFilterStringErrorsParsingEventArgsDescription;
})();
