// igniteui-react-core/esm2015/lib/DataGridSortIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataGridSortIndicatorDescription = /*@__PURE__*/ (() => {
    class DataGridSortIndicatorDescription extends Description {
        get_type() {
            return "DataGridSortIndicator";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.p = null;
            this.n = null;
            this.h = false;
            this.k = 0;
            this.j = 0;
            this.q = null;
            this.o = null;
        }
        get sortDirection() {
            return this.p;
        }
        set sortDirection(a) {
            this.p = a;
            this.g("SortDirection");
        }
        get iconColor() {
            return this.n;
        }
        set iconColor(a) {
            this.n = a;
            this.g("IconColor");
        }
        get isAnimationEnabled() {
            return this.h;
        }
        set isAnimationEnabled(a) {
            this.h = a;
            this.g("IsAnimationEnabled");
        }
        get pixelScalingRatio() {
            return this.k;
        }
        set pixelScalingRatio(a) {
            this.k = a;
            this.g("PixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.j;
        }
        set actualPixelScalingRatio(a) {
            this.j = a;
            this.g("ActualPixelScalingRatio");
        }
        get sortIndicatorStyle() {
            return this.q;
        }
        set sortIndicatorStyle(a) {
            this.q = a;
            this.g("SortIndicatorStyle");
        }
        get renderCompletedRef() {
            return this.o;
        }
        set renderCompletedRef(a) {
            this.o = a;
            this.g("RenderCompletedRef");
        }
    }
    DataGridSortIndicatorDescription.$t = /*@__PURE__*/ markType(DataGridSortIndicatorDescription, 'DataGridSortIndicatorDescription', Description.$);
    return DataGridSortIndicatorDescription;
})();
