// igniteui-react-core/esm2015/lib/DataLegendStylingRowEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataLegendStylingRowEventArgsDescription = /*@__PURE__*/ (() => {
    class DataLegendStylingRowEventArgsDescription extends Description {
        get_type() {
            return "DataLegendStylingRowEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.h = false;
            this.i = false;
            this.l = 0;
            this.p = null;
            this.q = null;
            this.o = null;
            this.r = null;
        }
        get badgeShape() {
            return this.n;
        }
        set badgeShape(a) {
            this.n = a;
            this.g("BadgeShape");
        }
        get isBadgeVisible() {
            return this.h;
        }
        set isBadgeVisible(a) {
            this.h = a;
            this.g("IsBadgeVisible");
        }
        get isRowVisible() {
            return this.i;
        }
        set isRowVisible(a) {
            this.i = a;
            this.g("IsRowVisible");
        }
        get seriesIndex() {
            return this.l;
        }
        set seriesIndex(a) {
            this.l = a;
            this.g("SeriesIndex");
        }
        get seriesTitle() {
            return this.p;
        }
        set seriesTitle(a) {
            this.p = a;
            this.g("SeriesTitle");
        }
        get titleText() {
            return this.q;
        }
        set titleText(a) {
            this.q = a;
            this.g("TitleText");
        }
        get groupName() {
            return this.o;
        }
        set groupName(a) {
            this.o = a;
            this.g("GroupName");
        }
        get titleTextColor() {
            return this.r;
        }
        set titleTextColor(a) {
            this.r = a;
            this.g("TitleTextColor");
        }
    }
    DataLegendStylingRowEventArgsDescription.$t = /*@__PURE__*/ markType(DataLegendStylingRowEventArgsDescription, 'DataLegendStylingRowEventArgsDescription', Description.$);
    return DataLegendStylingRowEventArgsDescription;
})();
