// igniteui-react-core/esm2015/lib/DataGridDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataGridDescription = /*@__PURE__*/ (() => {
    class DataGridDescription extends Description {
        get width() {
            return this.ix;
        }
        set width(a) {
            this.ix = a;
            this.g("Width");
        }
        get height() {
            return this.g2;
        }
        set height(a) {
            this.g2 = a;
            this.g("Height");
        }
        get background() {
            return this.e6;
        }
        set background(a) {
            this.e6 = a;
            this.g("Background");
        }
        get isGroupByAreaVisible() {
            return this.bt;
        }
        set isGroupByAreaVisible(a) {
            this.bt = a;
            this.g("IsGroupByAreaVisible");
        }
        get isPagerVisible() {
            return this.by;
        }
        set isPagerVisible(a) {
            this.by = a;
            this.g("IsPagerVisible");
        }
        get isToolbarVisible() {
            return this.b3;
        }
        set isToolbarVisible(a) {
            this.b3 = a;
            this.g("IsToolbarVisible");
        }
        get_type() {
            return "DataGrid";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.ix = null;
            this.g2 = null;
            this.e6 = null;
            this.bt = false;
            this.by = false;
            this.b3 = false;
            this.f1 = null;
            this.ft = null;
            this.b0 = false;
            this.hf = null;
            this.hg = null;
            this.gh = null;
            this.dh = 0;
            this.di = 0;
            this.dj = 0;
            this.dk = 0;
            this.bd = null;
            this.bf = null;
            this.bh = null;
            this.bb = null;
            this.a5 = null;
            this.a7 = null;
            this.az = null;
            this.ax = null;
            this.a9 = null;
            this.j = null;
            this.l = null;
            this.gb = null;
            this.ej = 0;
            this.eq = 0;
            this.de = 0;
            this.ei = 0;
            this.em = 0;
            this.f9 = null;
            this.av = null;
            this.gt = null;
            this.gf = null;
            this.gg = null;
            this.bo = false;
            this.bl = false;
            this.er = 0;
            this.bz = false;
            this.fo = null;
            this.fz = null;
            this.fa = null;
            this.hh = null;
            this.he = null;
            this.fl = null;
            this.a1 = null;
            this.f0 = null;
            this.f4 = null;
            this.fs = null;
            this.fr = null;
            this.fp = null;
            this.h0 = null;
            this.e2 = null;
            this.hy = null;
            this.b4 = false;
            this.x = null;
            this.y = null;
            this.w = null;
            this.v = null;
            this.r = null;
            this.s = null;
            this.bq = false;
            this.bw = false;
            this.bx = false;
            this.fk = null;
            this.hs = null;
            this.h5 = null;
            this.ih = null;
            this.dm = 0;
            this.gc = null;
            this.df = 0;
            this.gd = null;
            this.ge = null;
            this.dg = 0;
            this.ga = null;
            this.h3 = null;
            this.ha = null;
            this.g6 = null;
            this.a3 = null;
            this.e8 = null;
            this.fm = null;
            this.gz = null;
            this.g0 = null;
            this.fd = null;
            this.c7 = 0;
            this.fe = null;
            this.ff = null;
            this.ho = null;
            this.ht = null;
            this.hp = null;
            this.dn = 0;
            this.hq = null;
            this.hr = null;
            this.h6 = null;
            this.dp = 0;
            this.h7 = null;
            this.h8 = null;
            this.ii = null;
            this.dr = 0;
            this.ij = null;
            this.ik = null;
            this.ie = null;
            this.ib = null;
            this.dq = 0;
            this.ic = null;
            this.id = null;
            this.io = null;
            this.ds = 0;
            this.ip = null;
            this.iq = null;
            this.ir = null;
            this.h9 = null;
            this.il = null;
            this.ia = null;
            this.im = null;
            this.ep = 0;
            this.en = 0;
            this.gy = null;
            this.gs = null;
            this.g1 = null;
            this.gu = null;
            this.dl = 0;
            this.gv = null;
            this.gw = null;
            this.hi = null;
            this.gx = null;
            this.hl = null;
            this.hj = null;
            this.hk = null;
            this.el = 0;
            this.c9 = 0;
            this.ek = 0;
            this.c8 = 0;
            this.f2 = null;
            this.fu = null;
            this.b9 = false;
            this.fv = null;
            this.fx = null;
            this.fw = null;
            this.e7 = null;
            this.c3 = 0;
            this.c4 = 0;
            this.c5 = 0;
            this.c6 = 0;
            this.b5 = false;
            this.db = 0;
            this.da = 0;
            this.dc = 0;
            this.dd = 0;
            this.bp = false;
            this.u = null;
            this.p = null;
            this.o = null;
            this.q = null;
            this.g4 = null;
            this.g3 = null;
            this.g5 = null;
            this.z = null;
            this.n = null;
            this.m = null;
            this.aa = null;
            this.ig = null;
            this.bu = false;
            this.gq = null;
            this.bv = false;
            this.gr = null;
            this.bk = false;
            this.t = null;
            this.i = null;
            this.h = null;
            this.k = null;
            this.b8 = false;
            this.b7 = false;
            this.b6 = false;
            this.gi = null;
            this.gj = null;
            this.bs = false;
            this.br = false;
            this.g8 = null;
            this.g7 = null;
            this.g9 = null;
            this.go = null;
            this.gn = null;
            this.gk = null;
            this.c1 = new Date();
            this.bm = false;
            this.eo = 0;
            this.b1 = false;
            this.b2 = false;
            this.iw = null;
            this.is = null;
            this.it = null;
            this.iu = null;
            this.iv = null;
            this.hn = null;
            this.hm = null;
            this.bj = false;
            this.bn = false;
            this.gm = null;
            this.gl = null;
            this.fi = null;
            this.fj = null;
            this.fg = null;
            this.fh = null;
            this.e9 = null;
            this.h2 = null;
            this.gp = null;
            this.h4 = null;
            this.fq = null;
            this.fy = null;
            this.fn = null;
            this.f7 = null;
            this.f8 = null;
            this.hb = null;
            this.h1 = null;
            this.e4 = null;
            this.f3 = null;
            this.hw = null;
            this.hx = null;
            this.hv = null;
            this.hz = null;
            this.e3 = null;
            this.hu = null;
            this.f5 = null;
            this.fc = null;
            this.fb = null;
            this.hd = null;
            this.hc = null;
            this.f6 = null;
            this.e5 = null;
        }
        get columnResizingMode() {
            return this.f1;
        }
        set columnResizingMode(a) {
            this.f1 = a;
            this.g("ColumnResizingMode");
        }
        get columnMovingMode() {
            return this.ft;
        }
        set columnMovingMode(a) {
            this.ft = a;
            this.g("ColumnMovingMode");
        }
        get isRowHoverEnabled() {
            return this.b0;
        }
        set isRowHoverEnabled(a) {
            this.b0 = a;
            this.g("IsRowHoverEnabled");
        }
        get rowHoverBackground() {
            return this.hf;
        }
        set rowHoverBackground(a) {
            this.hf = a;
            this.g("RowHoverBackground");
        }
        get rowHoverTextColor() {
            return this.hg;
        }
        set rowHoverTextColor(a) {
            this.hg = a;
            this.g("RowHoverTextColor");
        }
        get editRowBorder() {
            return this.gh;
        }
        set editRowBorder(a) {
            this.gh = a;
            this.g("EditRowBorder");
        }
        get editRowBorderWidthBottom() {
            return this.dh;
        }
        set editRowBorderWidthBottom(a) {
            this.dh = a;
            this.g("EditRowBorderWidthBottom");
        }
        get editRowBorderWidthLeft() {
            return this.di;
        }
        set editRowBorderWidthLeft(a) {
            this.di = a;
            this.g("EditRowBorderWidthLeft");
        }
        get editRowBorderWidthRight() {
            return this.dj;
        }
        set editRowBorderWidthRight(a) {
            this.dj = a;
            this.g("EditRowBorderWidthRight");
        }
        get editRowBorderWidthTop() {
            return this.dk;
        }
        set editRowBorderWidthTop(a) {
            this.dk = a;
            this.g("EditRowBorderWidthTop");
        }
        get sectionHeader() {
            return this.bd;
        }
        set sectionHeader(a) {
            this.bd = a;
            this.g("SectionHeader");
        }
        get summaryRowRoot() {
            return this.bf;
        }
        set summaryRowRoot(a) {
            this.bf = a;
            this.g("SummaryRowRoot");
        }
        get summaryRowSection() {
            return this.bh;
        }
        set summaryRowSection(a) {
            this.bh = a;
            this.g("SummaryRowSection");
        }
        get rowSeparator() {
            return this.bb;
        }
        set rowSeparator(a) {
            this.bb = a;
            this.g("RowSeparator");
        }
        get headerRowSeparator() {
            return this.a5;
        }
        set headerRowSeparator(a) {
            this.a5 = a;
            this.g("HeaderRowSeparator");
        }
        get headerSeparator() {
            return this.a7;
        }
        set headerSeparator(a) {
            this.a7 = a;
            this.g("HeaderSeparator");
        }
        get columnResizingSeparator() {
            return this.az;
        }
        set columnResizingSeparator(a) {
            this.az = a;
            this.g("ColumnResizingSeparator");
        }
        get columnMovingSeparator() {
            return this.ax;
        }
        set columnMovingSeparator(a) {
            this.ax = a;
            this.g("ColumnMovingSeparator");
        }
        get pinnedAreaSeparator() {
            return this.a9;
        }
        set pinnedAreaSeparator(a) {
            this.a9 = a;
            this.g("PinnedAreaSeparator");
        }
        get autoGenerateDesiredProperties() {
            return this.j;
        }
        set autoGenerateDesiredProperties(a) {
            this.j = a;
            this.g("AutoGenerateDesiredProperties");
        }
        get dataSourceDesiredProperties() {
            return this.l;
        }
        set dataSourceDesiredProperties(a) {
            this.l = a;
            this.g("DataSourceDesiredProperties");
        }
        get density() {
            return this.gb;
        }
        set density(a) {
            this.gb = a;
            this.g("Density");
        }
        get actualRowHeight() {
            return this.ej;
        }
        set actualRowHeight(a) {
            this.ej = a;
            this.g("ActualRowHeight");
        }
        get rowHeight() {
            return this.eq;
        }
        set rowHeight(a) {
            this.eq = a;
            this.g("RowHeight");
        }
        get defaultColumnMinWidth() {
            return this.de;
        }
        set defaultColumnMinWidth(a) {
            this.de = a;
            this.g("DefaultColumnMinWidth");
        }
        get actualHeaderHeight() {
            return this.ei;
        }
        set actualHeaderHeight(a) {
            this.ei = a;
            this.g("ActualHeaderHeight");
        }
        get headerHeight() {
            return this.em;
        }
        set headerHeight(a) {
            this.em = a;
            this.g("HeaderHeight");
        }
        get dataSourceRef() {
            return this.f9;
        }
        set dataSourceRef(a) {
            this.f9 = a;
            this.g("DataSourceRef");
        }
        get dataSource() {
            return this.av;
        }
        set dataSource(a) {
            this.av = a;
            this.g("DataSource");
        }
        get headerClickAction() {
            return this.gt;
        }
        set headerClickAction(a) {
            this.gt = a;
            this.g("HeaderClickAction");
        }
        get editMode() {
            return this.gf;
        }
        set editMode(a) {
            this.gf = a;
            this.g("EditMode");
        }
        get editModeClickAction() {
            return this.gg;
        }
        set editModeClickAction(a) {
            this.gg = a;
            this.g("EditModeClickAction");
        }
        get editOnKeyPress() {
            return this.bo;
        }
        set editOnKeyPress(a) {
            this.bo = a;
            this.g("EditOnKeyPress");
        }
        get autoGenerateColumns() {
            return this.bl;
        }
        set autoGenerateColumns(a) {
            this.bl = a;
            this.g("AutoGenerateColumns");
        }
        get rowSeparatorHeight() {
            return this.er;
        }
        set rowSeparatorHeight(a) {
            this.er = a;
            this.g("RowSeparatorHeight");
        }
        get isPlaceholderRenderingEnabled() {
            return this.bz;
        }
        set isPlaceholderRenderingEnabled(a) {
            this.bz = a;
            this.g("IsPlaceholderRenderingEnabled");
        }
        get columnAddingAnimationMode() {
            return this.fo;
        }
        set columnAddingAnimationMode(a) {
            this.fo = a;
            this.g("ColumnAddingAnimationMode");
        }
        get columnPropertyUpdatingAnimationMode() {
            return this.fz;
        }
        set columnPropertyUpdatingAnimationMode(a) {
            this.fz = a;
            this.g("ColumnPropertyUpdatingAnimationMode");
        }
        get cellDataLoadedAnimationMode() {
            return this.fa;
        }
        set cellDataLoadedAnimationMode(a) {
            this.fa = a;
            this.g("CellDataLoadedAnimationMode");
        }
        get rowSelectionAnimationMode() {
            return this.hh;
        }
        set rowSelectionAnimationMode(a) {
            this.hh = a;
            this.g("RowSelectionAnimationMode");
        }
        get rowHoverAnimationMode() {
            return this.he;
        }
        set rowHoverAnimationMode(a) {
            this.he = a;
            this.g("RowHoverAnimationMode");
        }
        get cellSelectionAnimationMode() {
            return this.fl;
        }
        set cellSelectionAnimationMode(a) {
            this.fl = a;
            this.g("CellSelectionAnimationMode");
        }
        get animationSettings() {
            return this.a1;
        }
        set animationSettings(a) {
            this.a1 = a;
            this.g("AnimationSettings");
        }
        get columnResizingAnimationMode() {
            return this.f0;
        }
        set columnResizingAnimationMode(a) {
            this.f0 = a;
            this.g("ColumnResizingAnimationMode");
        }
        get columnShowingAnimationMode() {
            return this.f4;
        }
        set columnShowingAnimationMode(a) {
            this.f4 = a;
            this.g("ColumnShowingAnimationMode");
        }
        get columnMovingAnimationMode() {
            return this.fs;
        }
        set columnMovingAnimationMode(a) {
            this.fs = a;
            this.g("ColumnMovingAnimationMode");
        }
        get columnHidingAnimationMode() {
            return this.fr;
        }
        set columnHidingAnimationMode(a) {
            this.fr = a;
            this.g("ColumnHidingAnimationMode");
        }
        get columnExchangingAnimationMode() {
            return this.fp;
        }
        set columnExchangingAnimationMode(a) {
            this.fp = a;
            this.g("ColumnExchangingAnimationMode");
        }
        get selectionMode() {
            return this.h0;
        }
        set selectionMode(a) {
            this.h0 = a;
            this.g("SelectionMode");
        }
        get activationMode() {
            return this.e2;
        }
        set activationMode(a) {
            this.e2 = a;
            this.g("ActivationMode");
        }
        get selectionBehavior() {
            return this.hy;
        }
        set selectionBehavior(a) {
            this.hy = a;
            this.g("SelectionBehavior");
        }
        get mouseDragSelectionEnabled() {
            return this.b4;
        }
        set mouseDragSelectionEnabled(a) {
            this.b4 = a;
            this.g("MouseDragSelectionEnabled");
        }
        get selectedItems() {
            return this.x;
        }
        set selectedItems(a) {
            this.x = a;
            this.g("SelectedItems");
        }
        get selectedKeys() {
            return this.y;
        }
        set selectedKeys(a) {
            this.y = a;
            this.g("SelectedKeys");
        }
        get selectedCells() {
            return this.w;
        }
        set selectedCells(a) {
            this.w = a;
            this.g("SelectedCells");
        }
        get selectedCellRanges() {
            return this.v;
        }
        set selectedCellRanges(a) {
            this.v = a;
            this.g("SelectedCellRanges");
        }
        get pinnedItems() {
            return this.r;
        }
        set pinnedItems(a) {
            this.r = a;
            this.g("PinnedItems");
        }
        get pinnedKeys() {
            return this.s;
        }
        set pinnedKeys(a) {
            this.s = a;
            this.g("PinnedKeys");
        }
        get isColumnOptionsEnabled() {
            return this.bq;
        }
        set isColumnOptionsEnabled(a) {
            this.bq = a;
            this.g("IsColumnOptionsEnabled");
        }
        get isGroupRowSticky() {
            return this.bw;
        }
        set isGroupRowSticky(a) {
            this.bw = a;
            this.g("IsGroupRowSticky");
        }
        get isHeaderSeparatorVisible() {
            return this.bx;
        }
        set isHeaderSeparatorVisible(a) {
            this.bx = a;
            this.g("IsHeaderSeparatorVisible");
        }
        get cellSelectedBackground() {
            return this.fk;
        }
        set cellSelectedBackground(a) {
            this.fk = a;
            this.g("CellSelectedBackground");
        }
        get sectionHeaderSelectedBackground() {
            return this.hs;
        }
        set sectionHeaderSelectedBackground(a) {
            this.hs = a;
            this.g("SectionHeaderSelectedBackground");
        }
        get summaryRootBackground() {
            return this.h5;
        }
        set summaryRootBackground(a) {
            this.h5 = a;
            this.g("SummaryRootBackground");
        }
        get summarySectionBackground() {
            return this.ih;
        }
        set summarySectionBackground(a) {
            this.ih = a;
            this.g("SummarySectionBackground");
        }
        get pinnedRowOpacity() {
            return this.dm;
        }
        set pinnedRowOpacity(a) {
            this.dm = a;
            this.g("PinnedRowOpacity");
        }
        get editFontFamily() {
            return this.gc;
        }
        set editFontFamily(a) {
            this.gc = a;
            this.g("EditFontFamily");
        }
        get editFontSize() {
            return this.df;
        }
        set editFontSize(a) {
            this.df = a;
            this.g("EditFontSize");
        }
        get editFontStyle() {
            return this.gd;
        }
        set editFontStyle(a) {
            this.gd = a;
            this.g("EditFontStyle");
        }
        get editFontWeight() {
            return this.ge;
        }
        set editFontWeight(a) {
            this.ge = a;
            this.g("EditFontWeight");
        }
        get editOpacity() {
            return this.dg;
        }
        set editOpacity(a) {
            this.dg = a;
            this.g("EditOpacity");
        }
        get deletedTextColor() {
            return this.ga;
        }
        set deletedTextColor(a) {
            this.ga = a;
            this.g("DeletedTextColor");
        }
        get stickyRowBackground() {
            return this.h3;
        }
        set stickyRowBackground(a) {
            this.h3 = a;
            this.g("StickyRowBackground");
        }
        get pinnedRowBackground() {
            return this.ha;
        }
        set pinnedRowBackground(a) {
            this.ha = a;
            this.g("PinnedRowBackground");
        }
        get lastStickyRowBackground() {
            return this.g6;
        }
        set lastStickyRowBackground(a) {
            this.g6 = a;
            this.g("LastStickyRowBackground");
        }
        get activeCell() {
            return this.a3;
        }
        set activeCell(a) {
            this.a3 = a;
            this.g("ActiveCell");
        }
        get cellBackground() {
            return this.e8;
        }
        set cellBackground(a) {
            this.e8 = a;
            this.g("CellBackground");
        }
        get cellTextColor() {
            return this.fm;
        }
        set cellTextColor(a) {
            this.fm = a;
            this.g("CellTextColor");
        }
        get headerSortIndicatorColor() {
            return this.gz;
        }
        set headerSortIndicatorColor(a) {
            this.gz = a;
            this.g("HeaderSortIndicatorColor");
        }
        get headerSortIndicatorStyle() {
            return this.g0;
        }
        set headerSortIndicatorStyle(a) {
            this.g0 = a;
            this.g("HeaderSortIndicatorStyle");
        }
        get cellFontFamily() {
            return this.fd;
        }
        set cellFontFamily(a) {
            this.fd = a;
            this.g("CellFontFamily");
        }
        get cellFontSize() {
            return this.c7;
        }
        set cellFontSize(a) {
            this.c7 = a;
            this.g("CellFontSize");
        }
        get cellFontStyle() {
            return this.fe;
        }
        set cellFontStyle(a) {
            this.fe = a;
            this.g("CellFontStyle");
        }
        get cellFontWeight() {
            return this.ff;
        }
        set cellFontWeight(a) {
            this.ff = a;
            this.g("CellFontWeight");
        }
        get sectionHeaderBackground() {
            return this.ho;
        }
        set sectionHeaderBackground(a) {
            this.ho = a;
            this.g("SectionHeaderBackground");
        }
        get sectionHeaderTextColor() {
            return this.ht;
        }
        set sectionHeaderTextColor(a) {
            this.ht = a;
            this.g("SectionHeaderTextColor");
        }
        get sectionHeaderFontFamily() {
            return this.hp;
        }
        set sectionHeaderFontFamily(a) {
            this.hp = a;
            this.g("SectionHeaderFontFamily");
        }
        get sectionHeaderFontSize() {
            return this.dn;
        }
        set sectionHeaderFontSize(a) {
            this.dn = a;
            this.g("SectionHeaderFontSize");
        }
        get sectionHeaderFontStyle() {
            return this.hq;
        }
        set sectionHeaderFontStyle(a) {
            this.hq = a;
            this.g("SectionHeaderFontStyle");
        }
        get sectionHeaderFontWeight() {
            return this.hr;
        }
        set sectionHeaderFontWeight(a) {
            this.hr = a;
            this.g("SectionHeaderFontWeight");
        }
        get summaryRootLabelFontFamily() {
            return this.h6;
        }
        set summaryRootLabelFontFamily(a) {
            this.h6 = a;
            this.g("SummaryRootLabelFontFamily");
        }
        get summaryRootLabelFontSize() {
            return this.dp;
        }
        set summaryRootLabelFontSize(a) {
            this.dp = a;
            this.g("SummaryRootLabelFontSize");
        }
        get summaryRootLabelFontStyle() {
            return this.h7;
        }
        set summaryRootLabelFontStyle(a) {
            this.h7 = a;
            this.g("SummaryRootLabelFontStyle");
        }
        get summaryRootLabelFontWeight() {
            return this.h8;
        }
        set summaryRootLabelFontWeight(a) {
            this.h8 = a;
            this.g("SummaryRootLabelFontWeight");
        }
        get summarySectionLabelFontFamily() {
            return this.ii;
        }
        set summarySectionLabelFontFamily(a) {
            this.ii = a;
            this.g("SummarySectionLabelFontFamily");
        }
        get summarySectionLabelFontSize() {
            return this.dr;
        }
        set summarySectionLabelFontSize(a) {
            this.dr = a;
            this.g("SummarySectionLabelFontSize");
        }
        get summarySectionLabelFontStyle() {
            return this.ij;
        }
        set summarySectionLabelFontStyle(a) {
            this.ij = a;
            this.g("SummarySectionLabelFontStyle");
        }
        get summarySectionLabelFontWeight() {
            return this.ik;
        }
        set summarySectionLabelFontWeight(a) {
            this.ik = a;
            this.g("SummarySectionLabelFontWeight");
        }
        get summaryRootValueTextColor() {
            return this.ie;
        }
        set summaryRootValueTextColor(a) {
            this.ie = a;
            this.g("SummaryRootValueTextColor");
        }
        get summaryRootValueFontFamily() {
            return this.ib;
        }
        set summaryRootValueFontFamily(a) {
            this.ib = a;
            this.g("SummaryRootValueFontFamily");
        }
        get summaryRootValueFontSize() {
            return this.dq;
        }
        set summaryRootValueFontSize(a) {
            this.dq = a;
            this.g("SummaryRootValueFontSize");
        }
        get summaryRootValueFontStyle() {
            return this.ic;
        }
        set summaryRootValueFontStyle(a) {
            this.ic = a;
            this.g("SummaryRootValueFontStyle");
        }
        get summaryRootValueFontWeight() {
            return this.id;
        }
        set summaryRootValueFontWeight(a) {
            this.id = a;
            this.g("SummaryRootValueFontWeight");
        }
        get summarySectionValueFontFamily() {
            return this.io;
        }
        set summarySectionValueFontFamily(a) {
            this.io = a;
            this.g("SummarySectionValueFontFamily");
        }
        get summarySectionValueFontSize() {
            return this.ds;
        }
        set summarySectionValueFontSize(a) {
            this.ds = a;
            this.g("SummarySectionValueFontSize");
        }
        get summarySectionValueFontStyle() {
            return this.ip;
        }
        set summarySectionValueFontStyle(a) {
            this.ip = a;
            this.g("SummarySectionValueFontStyle");
        }
        get summarySectionValueFontWeight() {
            return this.iq;
        }
        set summarySectionValueFontWeight(a) {
            this.iq = a;
            this.g("SummarySectionValueFontWeight");
        }
        get summarySectionValueTextColor() {
            return this.ir;
        }
        set summarySectionValueTextColor(a) {
            this.ir = a;
            this.g("SummarySectionValueTextColor");
        }
        get summaryRootLabelTextColor() {
            return this.h9;
        }
        set summaryRootLabelTextColor(a) {
            this.h9 = a;
            this.g("SummaryRootLabelTextColor");
        }
        get summarySectionLabelTextColor() {
            return this.il;
        }
        set summarySectionLabelTextColor(a) {
            this.il = a;
            this.g("SummarySectionLabelTextColor");
        }
        get summaryRootSelectedBackground() {
            return this.ia;
        }
        set summaryRootSelectedBackground(a) {
            this.ia = a;
            this.g("SummaryRootSelectedBackground");
        }
        get summarySectionSelectedBackground() {
            return this.im;
        }
        set summarySectionSelectedBackground(a) {
            this.im = a;
            this.g("SummarySectionSelectedBackground");
        }
        get pinnedAreaSeparatorWidth() {
            return this.ep;
        }
        set pinnedAreaSeparatorWidth(a) {
            this.ep = a;
            this.g("PinnedAreaSeparatorWidth");
        }
        get headerSeparatorWidth() {
            return this.en;
        }
        set headerSeparatorWidth(a) {
            this.en = a;
            this.g("HeaderSeparatorWidth");
        }
        get headerSeparatorBackground() {
            return this.gy;
        }
        set headerSeparatorBackground(a) {
            this.gy = a;
            this.g("HeaderSeparatorBackground");
        }
        get headerBackground() {
            return this.gs;
        }
        set headerBackground(a) {
            this.gs = a;
            this.g("HeaderBackground");
        }
        get headerTextColor() {
            return this.g1;
        }
        set headerTextColor(a) {
            this.g1 = a;
            this.g("HeaderTextColor");
        }
        get headerFontFamily() {
            return this.gu;
        }
        set headerFontFamily(a) {
            this.gu = a;
            this.g("HeaderFontFamily");
        }
        get headerFontSize() {
            return this.dl;
        }
        set headerFontSize(a) {
            this.dl = a;
            this.g("HeaderFontSize");
        }
        get headerFontStyle() {
            return this.gv;
        }
        set headerFontStyle(a) {
            this.gv = a;
            this.g("HeaderFontStyle");
        }
        get headerFontWeight() {
            return this.gw;
        }
        set headerFontWeight(a) {
            this.gw = a;
            this.g("HeaderFontWeight");
        }
        get rowSeparatorBackground() {
            return this.hi;
        }
        set rowSeparatorBackground(a) {
            this.hi = a;
            this.g("RowSeparatorBackground");
        }
        get headerRowSeparatorBackground() {
            return this.gx;
        }
        set headerRowSeparatorBackground(a) {
            this.gx = a;
            this.g("HeaderRowSeparatorBackground");
        }
        get rowSeparatorStickyRowBackground() {
            return this.hl;
        }
        set rowSeparatorStickyRowBackground(a) {
            this.hl = a;
            this.g("RowSeparatorStickyRowBackground");
        }
        get rowSeparatorLastStickyRowBackground() {
            return this.hj;
        }
        set rowSeparatorLastStickyRowBackground(a) {
            this.hj = a;
            this.g("RowSeparatorLastStickyRowBackground");
        }
        get rowSeparatorPinnedRowBackground() {
            return this.hk;
        }
        set rowSeparatorPinnedRowBackground(a) {
            this.hk = a;
            this.g("RowSeparatorPinnedRowBackground");
        }
        get columnResizingSeparatorWidth() {
            return this.el;
        }
        set columnResizingSeparatorWidth(a) {
            this.el = a;
            this.g("ColumnResizingSeparatorWidth");
        }
        get columnResizingSeparatorOpacity() {
            return this.c9;
        }
        set columnResizingSeparatorOpacity(a) {
            this.c9 = a;
            this.g("ColumnResizingSeparatorOpacity");
        }
        get columnMovingSeparatorWidth() {
            return this.ek;
        }
        set columnMovingSeparatorWidth(a) {
            this.ek = a;
            this.g("ColumnMovingSeparatorWidth");
        }
        get columnMovingSeparatorOpacity() {
            return this.c8;
        }
        set columnMovingSeparatorOpacity(a) {
            this.c8 = a;
            this.g("ColumnMovingSeparatorOpacity");
        }
        get columnResizingSeparatorBackground() {
            return this.f2;
        }
        set columnResizingSeparatorBackground(a) {
            this.f2 = a;
            this.g("ColumnResizingSeparatorBackground");
        }
        get columnMovingSeparatorBackground() {
            return this.fu;
        }
        set columnMovingSeparatorBackground(a) {
            this.fu = a;
            this.g("ColumnMovingSeparatorBackground");
        }
        get shiftSectionContent() {
            return this.b9;
        }
        set shiftSectionContent(a) {
            this.b9 = a;
            this.g("ShiftSectionContent");
        }
        get columnOptionsIconAlignment() {
            return this.fv;
        }
        set columnOptionsIconAlignment(a) {
            this.fv = a;
            this.g("ColumnOptionsIconAlignment");
        }
        get columnOptionsIconColor() {
            return this.fx;
        }
        set columnOptionsIconColor(a) {
            this.fx = a;
            this.g("ColumnOptionsIconColor");
        }
        get columnOptionsIconBehavior() {
            return this.fw;
        }
        set columnOptionsIconBehavior(a) {
            this.fw = a;
            this.g("ColumnOptionsIconBehavior");
        }
        get border() {
            return this.e7;
        }
        set border(a) {
            this.e7 = a;
            this.g("Border");
        }
        get borderWidthBottom() {
            return this.c3;
        }
        set borderWidthBottom(a) {
            this.c3 = a;
            this.g("BorderWidthBottom");
        }
        get borderWidthLeft() {
            return this.c4;
        }
        set borderWidthLeft(a) {
            this.c4 = a;
            this.g("BorderWidthLeft");
        }
        get borderWidthRight() {
            return this.c5;
        }
        set borderWidthRight(a) {
            this.c5 = a;
            this.g("BorderWidthRight");
        }
        get borderWidthTop() {
            return this.c6;
        }
        set borderWidthTop(a) {
            this.c6 = a;
            this.g("BorderWidthTop");
        }
        get notifyOnAllSelectionChanges() {
            return this.b5;
        }
        set notifyOnAllSelectionChanges(a) {
            this.b5 = a;
            this.g("NotifyOnAllSelectionChanges");
        }
        get cornerRadiusBottomRight() {
            return this.db;
        }
        set cornerRadiusBottomRight(a) {
            this.db = a;
            this.g("CornerRadiusBottomRight");
        }
        get cornerRadiusBottomLeft() {
            return this.da;
        }
        set cornerRadiusBottomLeft(a) {
            this.da = a;
            this.g("CornerRadiusBottomLeft");
        }
        get cornerRadiusTopLeft() {
            return this.dc;
        }
        set cornerRadiusTopLeft(a) {
            this.dc = a;
            this.g("CornerRadiusTopLeft");
        }
        get cornerRadiusTopRight() {
            return this.dd;
        }
        set cornerRadiusTopRight(a) {
            this.dd = a;
            this.g("CornerRadiusTopRight");
        }
        get isActiveCellStyleEnabled() {
            return this.bp;
        }
        set isActiveCellStyleEnabled(a) {
            this.bp = a;
            this.g("IsActiveCellStyleEnabled");
        }
        get responsiveStates() {
            return this.u;
        }
        set responsiveStates(a) {
            this.u = a;
            this.g("ResponsiveStates");
        }
        get initialSortDescriptions() {
            return this.p;
        }
        set initialSortDescriptions(a) {
            this.p = a;
            this.g("InitialSortDescriptions");
        }
        get initialGroupDescriptions() {
            return this.o;
        }
        set initialGroupDescriptions(a) {
            this.o = a;
            this.g("InitialGroupDescriptions");
        }
        get initialSummaryDescriptions() {
            return this.q;
        }
        set initialSummaryDescriptions(a) {
            this.q = a;
            this.g("InitialSummaryDescriptions");
        }
        get initialSorts() {
            return this.g4;
        }
        set initialSorts(a) {
            this.g4 = a;
            this.g("InitialSorts");
        }
        get initialGroups() {
            return this.g3;
        }
        set initialGroups(a) {
            this.g3 = a;
            this.g("InitialGroups");
        }
        get initialSummaries() {
            return this.g5;
        }
        set initialSummaries(a) {
            this.g5 = a;
            this.g("InitialSummaries");
        }
        get sortDescriptions() {
            return this.z;
        }
        set sortDescriptions(a) {
            this.z = a;
            this.g("SortDescriptions");
        }
        get groupDescriptions() {
            return this.n;
        }
        set groupDescriptions(a) {
            this.n = a;
            this.g("GroupDescriptions");
        }
        get filterExpressions() {
            return this.m;
        }
        set filterExpressions(a) {
            this.m = a;
            this.g("FilterExpressions");
        }
        get summaryDescriptions() {
            return this.aa;
        }
        set summaryDescriptions(a) {
            this.aa = a;
            this.g("SummaryDescriptions");
        }
        get summaryScope() {
            return this.ig;
        }
        set summaryScope(a) {
            this.ig = a;
            this.g("SummaryScope");
        }
        get isGroupCollapsable() {
            return this.bu;
        }
        set isGroupCollapsable(a) {
            this.bu = a;
            this.g("IsGroupCollapsable");
        }
        get groupHeaderDisplayMode() {
            return this.gq;
        }
        set groupHeaderDisplayMode(a) {
            this.gq = a;
            this.g("GroupHeaderDisplayMode");
        }
        get isGroupExpandedDefault() {
            return this.bv;
        }
        set isGroupExpandedDefault(a) {
            this.bv = a;
            this.g("IsGroupExpandedDefault");
        }
        get groupSummaryDisplayMode() {
            return this.gr;
        }
        set groupSummaryDisplayMode(a) {
            this.gr = a;
            this.g("GroupSummaryDisplayMode");
        }
        get autoAcceptEdits() {
            return this.bk;
        }
        set autoAcceptEdits(a) {
            this.bk = a;
            this.g("AutoAcceptEdits");
        }
        get primaryKey() {
            return this.t;
        }
        set primaryKey(a) {
            this.t = a;
            this.g("PrimaryKey");
        }
        get actualPrimaryKey() {
            return this.i;
        }
        set actualPrimaryKey(a) {
            this.i = a;
            this.g("ActualPrimaryKey");
        }
        get actualColumns() {
            return this.h;
        }
        set actualColumns(a) {
            this.h = a;
            this.g("ActualColumns");
        }
        get columns() {
            return this.k;
        }
        set columns(a) {
            this.k = a;
            this.g("Columns");
        }
        get reactsToSortChanges() {
            return this.b8;
        }
        set reactsToSortChanges(a) {
            this.b8 = a;
            this.g("ReactsToSortChanges");
        }
        get reactsToGroupChanges() {
            return this.b7;
        }
        set reactsToGroupChanges(a) {
            this.b7 = a;
            this.g("ReactsToGroupChanges");
        }
        get reactsToFilterChanges() {
            return this.b6;
        }
        set reactsToFilterChanges(a) {
            this.b6 = a;
            this.g("ReactsToFilterChanges");
        }
        get enterBehavior() {
            return this.gi;
        }
        set enterBehavior(a) {
            this.gi = a;
            this.g("EnterBehavior");
        }
        get enterBehaviorAfterEdit() {
            return this.gj;
        }
        set enterBehaviorAfterEdit(a) {
            this.gj = a;
            this.g("EnterBehaviorAfterEdit");
        }
        get isColumnOptionsSummariesEnabled() {
            return this.bs;
        }
        set isColumnOptionsSummariesEnabled(a) {
            this.bs = a;
            this.g("IsColumnOptionsSummariesEnabled");
        }
        get isColumnOptionsGroupingEnabled() {
            return this.br;
        }
        set isColumnOptionsGroupingEnabled(a) {
            this.br = a;
            this.g("IsColumnOptionsGroupingEnabled");
        }
        get mergedCellMode() {
            return this.g8;
        }
        set mergedCellMode(a) {
            this.g8 = a;
            this.g("MergedCellMode");
        }
        get mergedCellEvaluationCriteria() {
            return this.g7;
        }
        set mergedCellEvaluationCriteria(a) {
            this.g7 = a;
            this.g("MergedCellEvaluationCriteria");
        }
        get mergedCellVerticalAlignment() {
            return this.g9;
        }
        set mergedCellVerticalAlignment(a) {
            this.g9 = a;
            this.g("MergedCellVerticalAlignment");
        }
        get filterUIType() {
            return this.go;
        }
        set filterUIType(a) {
            this.go = a;
            this.g("FilterUIType");
        }
        get filterLogicalOperator() {
            return this.gn;
        }
        set filterLogicalOperator(a) {
            this.gn = a;
            this.g("FilterLogicalOperator");
        }
        get filterComparisonType() {
            return this.gk;
        }
        set filterComparisonType(a) {
            this.gk = a;
            this.g("FilterComparisonType");
        }
        get todayOverride() {
            return this.c1;
        }
        set todayOverride(a) {
            this.c1 = a;
            this.g("TodayOverride");
        }
        get clearSelectionOnEscape() {
            return this.bm;
        }
        set clearSelectionOnEscape(a) {
            this.bm = a;
            this.g("ClearSelectionOnEscape");
        }
        get pageSize() {
            return this.eo;
        }
        set pageSize(a) {
            this.eo = a;
            this.g("PageSize");
        }
        get isToolbarColumnChooserVisible() {
            return this.b1;
        }
        set isToolbarColumnChooserVisible(a) {
            this.b1 = a;
            this.g("IsToolbarColumnChooserVisible");
        }
        get isToolbarColumnPinningVisible() {
            return this.b2;
        }
        set isToolbarColumnPinningVisible(a) {
            this.b2 = a;
            this.g("IsToolbarColumnPinningVisible");
        }
        get toolbarTitle() {
            return this.iw;
        }
        set toolbarTitle(a) {
            this.iw = a;
            this.g("ToolbarTitle");
        }
        get toolbarColumnChooserText() {
            return this.is;
        }
        set toolbarColumnChooserText(a) {
            this.is = a;
            this.g("ToolbarColumnChooserText");
        }
        get toolbarColumnChooserTitle() {
            return this.it;
        }
        set toolbarColumnChooserTitle(a) {
            this.it = a;
            this.g("ToolbarColumnChooserTitle");
        }
        get toolbarColumnPinningText() {
            return this.iu;
        }
        set toolbarColumnPinningText(a) {
            this.iu = a;
            this.g("ToolbarColumnPinningText");
        }
        get toolbarColumnPinningTitle() {
            return this.iv;
        }
        set toolbarColumnPinningTitle(a) {
            this.iv = a;
            this.g("ToolbarColumnPinningTitle");
        }
        get scrollbarStyle() {
            return this.hn;
        }
        set scrollbarStyle(a) {
            this.hn = a;
            this.g("ScrollbarStyle");
        }
        get scrollbarBackground() {
            return this.hm;
        }
        set scrollbarBackground(a) {
            this.hm = a;
            this.g("ScrollbarBackground");
        }
        get allowCopyOperation() {
            return this.bj;
        }
        set allowCopyOperation(a) {
            this.bj = a;
            this.g("AllowCopyOperation");
        }
        get deferEventForRowDragSelection() {
            return this.bn;
        }
        set deferEventForRowDragSelection(a) {
            this.bn = a;
            this.g("DeferEventForRowDragSelection");
        }
        get filterExpressionsChangingRef() {
            return this.gm;
        }
        set filterExpressionsChangingRef(a) {
            this.gm = a;
            this.g("FilterExpressionsChangingRef");
        }
        get filterExpressionsChangedRef() {
            return this.gl;
        }
        set filterExpressionsChangedRef(a) {
            this.gl = a;
            this.g("FilterExpressionsChangedRef");
        }
        get cellPreviewPointerDownRef() {
            return this.fi;
        }
        set cellPreviewPointerDownRef(a) {
            this.fi = a;
            this.g("CellPreviewPointerDownRef");
        }
        get cellPreviewPointerUpRef() {
            return this.fj;
        }
        set cellPreviewPointerUpRef(a) {
            this.fj = a;
            this.g("CellPreviewPointerUpRef");
        }
        get cellPointerDownRef() {
            return this.fg;
        }
        set cellPointerDownRef(a) {
            this.fg = a;
            this.g("CellPointerDownRef");
        }
        get cellPointerUpRef() {
            return this.fh;
        }
        set cellPointerUpRef(a) {
            this.fh = a;
            this.g("CellPointerUpRef");
        }
        get cellClickedRef() {
            return this.e9;
        }
        set cellClickedRef(a) {
            this.e9 = a;
            this.g("CellClickedRef");
        }
        get sortDescriptionsChangedRef() {
            return this.h2;
        }
        set sortDescriptionsChangedRef(a) {
            this.h2 = a;
            this.g("SortDescriptionsChangedRef");
        }
        get groupDescriptionsChangedRef() {
            return this.gp;
        }
        set groupDescriptionsChangedRef(a) {
            this.gp = a;
            this.g("GroupDescriptionsChangedRef");
        }
        get summaryDescriptionsChangedRef() {
            return this.h4;
        }
        set summaryDescriptionsChangedRef(a) {
            this.h4 = a;
            this.g("SummaryDescriptionsChangedRef");
        }
        get columnHiddenChangedRef() {
            return this.fq;
        }
        set columnHiddenChangedRef(a) {
            this.fq = a;
            this.g("ColumnHiddenChangedRef");
        }
        get columnPinnedChangedRef() {
            return this.fy;
        }
        set columnPinnedChangedRef(a) {
            this.fy = a;
            this.g("ColumnPinnedChangedRef");
        }
        get cellValueChangingRef() {
            return this.fn;
        }
        set cellValueChangingRef(a) {
            this.fn = a;
            this.g("CellValueChangingRef");
        }
        get dataCommittedRef() {
            return this.f7;
        }
        set dataCommittedRef(a) {
            this.f7 = a;
            this.g("DataCommittedRef");
        }
        get dataCommittingRef() {
            return this.f8;
        }
        set dataCommittingRef(a) {
            this.f8 = a;
            this.g("DataCommittingRef");
        }
        get rootSummariesChangedRef() {
            return this.hb;
        }
        set rootSummariesChangedRef(a) {
            this.hb = a;
            this.g("RootSummariesChangedRef");
        }
        get sizeChangedRef() {
            return this.h1;
        }
        set sizeChangedRef(a) {
            this.h1 = a;
            this.g("SizeChangedRef");
        }
        get actualColumnsChangedRef() {
            return this.e4;
        }
        set actualColumnsChangedRef(a) {
            this.e4 = a;
            this.g("ActualColumnsChangedRef");
        }
        get columnsAutoGeneratedRef() {
            return this.f3;
        }
        set columnsAutoGeneratedRef(a) {
            this.f3 = a;
            this.g("ColumnsAutoGeneratedRef");
        }
        get selectedItemsChangedRef() {
            return this.hw;
        }
        set selectedItemsChangedRef(a) {
            this.hw = a;
            this.g("SelectedItemsChangedRef");
        }
        get selectedKeysChangedRef() {
            return this.hx;
        }
        set selectedKeysChangedRef(a) {
            this.hx = a;
            this.g("SelectedKeysChangedRef");
        }
        get selectedCellsChangedRef() {
            return this.hv;
        }
        set selectedCellsChangedRef(a) {
            this.hv = a;
            this.g("SelectedCellsChangedRef");
        }
        get selectionChangedRef() {
            return this.hz;
        }
        set selectionChangedRef(a) {
            this.hz = a;
            this.g("SelectionChangedRef");
        }
        get activeCellChangedRef() {
            return this.e3;
        }
        set activeCellChangedRef(a) {
            this.e3 = a;
            this.g("ActiveCellChangedRef");
        }
        get selectedCellRangesChangedRef() {
            return this.hu;
        }
        set selectedCellRangesChangedRef(a) {
            this.hu = a;
            this.g("SelectedCellRangesChangedRef");
        }
        get columnWidthChangedRef() {
            return this.f5;
        }
        set columnWidthChangedRef(a) {
            this.f5 = a;
            this.g("ColumnWidthChangedRef");
        }
        get cellEditStartedRef() {
            return this.fc;
        }
        set cellEditStartedRef(a) {
            this.fc = a;
            this.g("CellEditStartedRef");
        }
        get cellEditEndedRef() {
            return this.fb;
        }
        set cellEditEndedRef(a) {
            this.fb = a;
            this.g("CellEditEndedRef");
        }
        get rowEditStartedRef() {
            return this.hd;
        }
        set rowEditStartedRef(a) {
            this.hd = a;
            this.g("RowEditStartedRef");
        }
        get rowEditEndedRef() {
            return this.hc;
        }
        set rowEditEndedRef(a) {
            this.hc = a;
            this.g("RowEditEndedRef");
        }
        get customFilterRequestedRef() {
            return this.f6;
        }
        set customFilterRequestedRef(a) {
            this.f6 = a;
            this.g("CustomFilterRequestedRef");
        }
        get actualPrimaryKeyChangeRef() {
            return this.e5;
        }
        set actualPrimaryKeyChangeRef(a) {
            this.e5 = a;
            this.g("ActualPrimaryKeyChangeRef");
        }
    }
    DataGridDescription.$t = /*@__PURE__*/ markType(DataGridDescription, 'DataGridDescription', Description.$);
    return DataGridDescription;
})();
