// igniteui-react-core/esm2015/lib/DataTooltipGroupedPositionY.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Describes available positions on the Y axis at which the summary tooltip to be displayed.
 */
export var DataTooltipGroupedPositionY = /*@__PURE__*/ (function (DataTooltipGroupedPositionY) {
    /**
     * A position on the Y axis is selected automatically.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["Auto"] = 0] = "Auto";
    /**
     * Aligns the tooltip so that it appears at the top of the vertical snap point of the current series if applicable.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["SnapTop"] = 1] = "SnapTop";
    /**
     * Aligns the tooltip so that it appears at the bottom of the vertical snap point of the current series if applicable.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["SnapBottom"] = 2] = "SnapBottom";
    /**
     * Aligns the tooltip so that it appears at the middle of the vertical snap point of the current series if applicable.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["SnapMiddle"] = 3] = "SnapMiddle";
    /**
     * Aligns the tooltip so that it appears at the top of the vertical snap point of the current series if applicable.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["TopEdgeSnapTop"] = 4] = "TopEdgeSnapTop";
    /**
     * Aligns the tooltip so that it appears at the bottom of the vertical snap point of the current series if applicable.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["TopEdgeSnapBottom"] = 5] = "TopEdgeSnapBottom";
    /**
     * Aligns the tooltip so that it appears at the middle of the vertical snap point of the current series if applicable.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["TopEdgeSnapMiddle"] = 6] = "TopEdgeSnapMiddle";
    /**
     * Aligns the tooltip so that it appears at the top of the vertical snap point of the current series if applicable.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["BottomEdgeSnapTop"] = 7] = "BottomEdgeSnapTop";
    /**
     * Aligns the tooltip so that it appears at the bottom of the vertical snap point of the current series if applicable.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["BottomEdgeSnapBottom"] = 8] = "BottomEdgeSnapBottom";
    /**
     * Aligns the tooltip so that it appears at the middle of the vertical snap point of the current series if applicable.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["BottomEdgeSnapMiddle"] = 9] = "BottomEdgeSnapMiddle";
    /**
     * Aligns the tooltip so that it appears at the top of the cursor.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["TrackTop"] = 10] = "TrackTop";
    /**
     * Aligns the tooltip so that it appears at the bottom of the cursor.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["TrackBottom"] = 11] = "TrackBottom";
    /**
     * Aligns the tooltip so that the cursor appears in the center of the tooltip height.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["TrackMiddle"] = 12] = "TrackMiddle";
    /**
     * Aligns the tooltip relative to the top chart edge.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["PinTop"] = 13] = "PinTop";
    /**
     * Aligns the tooltip relative to the bottom chart edge.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["PinBottom"] = 14] = "PinBottom";
    /**
     * Aligns the tooltip height center with the height center of the chart.
     */
    DataTooltipGroupedPositionY[DataTooltipGroupedPositionY["PinMiddle"] = 15] = "PinMiddle";
    return DataTooltipGroupedPositionY;
})({});
/**
 * @hidden
 */
export let DataTooltipGroupedPositionY_$type = /*@__PURE__*/ markEnum('DataTooltipGroupedPositionY', 'Auto,0|SnapTop,1|SnapBottom,2|SnapMiddle,3|TopEdgeSnapTop,4|TopEdgeSnapBottom,5|TopEdgeSnapMiddle,6|BottomEdgeSnapTop,7|BottomEdgeSnapBottom,8|BottomEdgeSnapMiddle,9|TrackTop,10|TrackBottom,11|TrackMiddle,12|PinTop,13|PinBottom,14|PinMiddle,15');
