// igniteui-react-core/esm2015/lib/DataGridExpansionIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataGridExpansionIndicatorDescription = /*@__PURE__*/ (() => {
    class DataGridExpansionIndicatorDescription extends Description {
        get_type() {
            return "DataGridExpansionIndicator";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = false;
            this.p = null;
            this.h = false;
            this.m = 0;
            this.l = 0;
        }
        get isExpanded() {
            return this.i;
        }
        set isExpanded(a) {
            this.i = a;
            this.g("IsExpanded");
        }
        get iconColor() {
            return this.p;
        }
        set iconColor(a) {
            this.p = a;
            this.g("IconColor");
        }
        get isAnimationEnabled() {
            return this.h;
        }
        set isAnimationEnabled(a) {
            this.h = a;
            this.g("IsAnimationEnabled");
        }
        get pixelScalingRatio() {
            return this.m;
        }
        set pixelScalingRatio(a) {
            this.m = a;
            this.g("PixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.l;
        }
        set actualPixelScalingRatio(a) {
            this.l = a;
            this.g("ActualPixelScalingRatio");
        }
    }
    DataGridExpansionIndicatorDescription.$t = /*@__PURE__*/ markType(DataGridExpansionIndicatorDescription, 'DataGridExpansionIndicatorDescription', Description.$);
    return DataGridExpansionIndicatorDescription;
})();
