// igniteui-react-core/esm2015/lib/DataBindingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataBindingEventArgsDescription = /*@__PURE__*/ (() => {
    class DataBindingEventArgsDescription extends Description {
        get_type() {
            return "DataBindingEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.i = null;
        }
        get resolvedValueRef() {
            return this.h;
        }
        set resolvedValueRef(a) {
            this.h = a;
            this.g("ResolvedValueRef");
        }
        get rowObjectRef() {
            return this.i;
        }
        set rowObjectRef(a) {
            this.i = a;
            this.g("RowObjectRef");
        }
    }
    DataBindingEventArgsDescription.$t = /*@__PURE__*/ markType(DataBindingEventArgsDescription, 'DataBindingEventArgsDescription', Description.$);
    return DataBindingEventArgsDescription;
})();
