// igniteui-react-core/esm2015/lib/DataVisualizationLocaleNl.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
export class DataVisualizationLocaleNl {
    constructor() {
        this["AssemblyNotIncluded"] = "Kan {0}.dll niet laden. U moet ervoor zorgen dat deze assembly is opgenomen in de XAP die op de Windows Phone is geïmplementeerd, bijvoorbeeld door er vanuit al uw projecten een verwijzing naar toe te voegen.";
        this["BreakEvenTitle_BreakEven"] = "Break-even";
        this["BreakEvenTitle_FixedCost"] = "Vaste kosten";
        this["BreakEvenTitle_LossArea"] = "Verliesgebied";
        this["BreakEvenTitle_MarginalProfit"] = "Marginale winst";
        this["BreakEvenTitle_ProfitArea"] = "Winstgebied";
        this["BreakEvenTitle_SafetyMargin"] = "Veiligheidsmarge";
        this["BreakEvenTitle_SalesRevenue"] = "Verkoopopbrengst";
        this["BreakEvenTitle_TotalCost"] = "Totale kosten";
        this["BreakEvenTitle_VariableCost"] = "Variabele kosten";
        this["BubbleSeries_Radius"] = "Radius";
        this["DataChart_InteractivityNotLoaded"] = "Er is toegang geweest tot de interactiviteitsfunctie, maar de grafiekinteractiemodule is niet geladen.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Er werd een functie voor getalafkorting gebruikt, maar de getalafkortingsmodule is niet geladen.";
        this["DataChart_VisualDataNotLoaded"] = "Er is toegang geweest tot de VisualData-functie, maar de grafiekinteractiemodule is niet geladen.";
        this["DataProviderNotInitialized"] = "Gegevensprovider is gebruikt zonder eerst te initialiseren. De gegevensprovider moet vóór gebruik worden geïnitialiseerd.";
        this["DataSource_Summary_Avg"] = "Gem";
        this["DataSource_Summary_Count"] = "Aant";
        this["DataSource_Summary_Max"] = "Max";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Som";
        this["Default_Series_Title"] = "Serietitel";
        this["FinancialChart_IndicatorMenu_Header"] = "Indicatoren";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Standalone indicatoren";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Overlays";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Trendlijnen";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volumes";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Oscillator met absoluut volume";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Accumulatie/Distributie";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Gemiddelde richtingindex";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Gemiddeld werkelijk bereik";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Bollinger-bandbreedte";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Chaikin-oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Chaikin-volatiliteit";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Commodity Channel Index";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Prijsoscillator zonder trend";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Ease Of Movement";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Snelle stochastische oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Krachtindex";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Volledige stochastische oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Market Facilitation Index";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Mass Index";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Mediane prijs";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Money Flow Index";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Negative Volume Index";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "On Balance Volume";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Percentage prijsoscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Percentage volumeoscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Positive Volume Index";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Prijsvolumetrend";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Langzame stochastische oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Standaardafwijking";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Normale prijs";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Ultieme oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Gewogen sluiten";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bollinger-banden";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Prijskanaal";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Derdegraads passend";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Cumulatief gemiddelde";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Exponentieel gemiddelde";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Exponentieel passend";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Lineair passend";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Logaritmisch passend";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Gewijzigd gemiddelde";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "Machtswet passend";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Kwadratisch passend";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Vierdegraads passend";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Vijfdegraads passend";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Eenvoudig gemiddelde";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Gewogen gemiddelde";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Gebied";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Kolom";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Lijn";
        this["FinancialChart_RangeSelector_All"] = "Alles";
        this["FinancialChart_RangeSelector_From"] = "Van";
        this["FinancialChart_RangeSelector_OneMonth"] = "1M";
        this["FinancialChart_RangeSelector_OneYear"] = "1J";
        this["FinancialChart_RangeSelector_SixMonths"] = "6M";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3M";
        this["FinancialChart_RangeSelector_To"] = "Tot";
        this["FinancialChart_RangeSelector_YearToDate"] = "JTH";
        this["FinancialSeries_Close"] = "Sluiten";
        this["FinancialSeries_High"] = "Hoog";
        this["FinancialSeries_Low"] = "Laag";
        this["FinancialSeries_Open"] = "Open";
        this["FinancialSeries_Volume"] = "Volume";
        this["NoEncodingsLoaded"] = "Er zijn geen coderingen gevonden! Laad de assemblage Infragistics.Encoding of bekijk de documentatie voor het maken van aangepaste coderingen.";
        this["NotSupportedEncoding"] = "De codering die overeenkomt met {0} {1} wordt niet ondersteund.";
        this["Object_Sealed"] = "Wijzigingen zijn niet toegestaan nadat het object is verzegeld.";
        this["OPD_DefaultInteraction"] = "Selecteer de standaard interactie voor slepen met de muis";
        this["OPD_DragPan"] = "Pan het oppervlak";
        this["OPD_DragZoom"] = "Inzoomen op selectiegebied";
        this["OPD_ScaleToFit"] = "Aanpassen aan scherm";
        this["OPD_ScaleToFit_SeriesViewer"] = "Zoomen opnieuw instellen (100%)";
        this["OPD_ZoomIn"] = "Inzoomen";
        this["OPD_ZoomOut"] = "Uitzoomen";
        this["OPD_ZoomTo100"] = "Zoomen naar 100%";
        this["PieChart_Others"] = "Overige";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Bereikgebaseerde wijzigingen van de gegevensbron worden momenteel niet ondersteund";
        this["ScatterSeries_Value"] = "Waarde";
        this["TRIAL_VERSION"] = "PROEF ZONDER LICENTIE Koop op Infragistics.com";
    }
}
