// igniteui-react-core/esm2015/lib/DataVisualizationLocaleJa.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
export class DataVisualizationLocaleJa {
    constructor() {
        this["AssemblyNotIncluded"] = "{0}.dll を読み込めません。このアセンブリが Windows Phone にデプロイされる XAP に含まれることを確認してください (プロジェクトから参照を追加など)。";
        this["BreakEvenTitle_BreakEven"] = "損益分岐点";
        this["BreakEvenTitle_FixedCost"] = "固定費";
        this["BreakEvenTitle_LossArea"] = "損失領域";
        this["BreakEvenTitle_MarginalProfit"] = "限界利益";
        this["BreakEvenTitle_ProfitArea"] = "利益領域";
        this["BreakEvenTitle_SafetyMargin"] = "安全余裕率";
        this["BreakEvenTitle_SalesRevenue"] = "売上収益";
        this["BreakEvenTitle_TotalCost"] = "総費用";
        this["BreakEvenTitle_VariableCost"] = "変動費";
        this["BubbleSeries_Radius"] = "半径";
        this["DataChart_InteractivityNotLoaded"] = "インタラクティブ機能にアクセスがありましたがインタラクション モジュールが読み込まれていません。";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "数値省略機能が使用されましたが、数値省略モジュールが読み込まれていません。";
        this["DataChart_VisualDataNotLoaded"] = "VisualData 機能にアクセスがありましたがインタラクション モジュールが読み込まれていません。";
        this["DataProviderNotInitialized"] = "初期化されていないデータ プロバイダーが使用されました。データ プロバイダーは使用前に初期化する必要があります。";
        this["DataSource_Summary_Avg"] = "Avg";
        this["DataSource_Summary_Count"] = "Count";
        this["DataSource_Summary_Max"] = "Max";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Sum";
        this["Default_Series_Title"] = "シリーズ タイトル";
        this["FinancialChart_IndicatorMenu_Header"] = "インジケーター";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "スタンドアロン インジケーター";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "オーバーレイ";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "トレンドライン";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "ボリューム";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "絶対出来高オシレーター";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "蓄積/分配";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "平均方向性指数";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "ATR (アベレージ トゥルー レンジ)";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "ボリンジャー バンド幅";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "チャイキン オシレーター";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "チャイキン ボラティリティ";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "商品チャネル指数 (CCI)";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "トレンド除去価格オシレーター";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "イーズ オブ ムーブメント";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "ファスト スト キャスティクス オシレーター";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "フォース インデックス";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "フル ストキャスティクス オシレーター";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "マーケット ファシリテーション インデックス";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "マス インデックス";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "メディアンプライス";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "マネー フロー インデックス";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "負出来高指数 (NVI)";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "OBV (オン バランス ボリューム) インジケーター";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "パーセンテージ価格オシレーター";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "パーセンテージ出来高オシレーター";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "PVI (ポジティブ ボリューム インデックス)";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "価格出来高トレンド";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "スロー ストキャスティクス オシレーター";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "標準偏差";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "ストキャスティクス RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "代表価格";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "アルティメット オシレーター";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "加重終値";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "ウィリアム パーセント";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "ボリンジャー バンド";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "プライス チャネル";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "キュービック フィット";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "累加平均";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "指数平均";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "指数フィット";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "リニア フィット";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "対数フィット";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "修正平均";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "PowerLaw フィット";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "二次フィット";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "四次フィット";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "五次フィット";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "単純平均";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "加重平均";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "エリア";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "柱状";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "折れ線";
        this["FinancialChart_RangeSelector_All"] = "すべて";
        this["FinancialChart_RangeSelector_From"] = "開始:";
        this["FinancialChart_RangeSelector_OneMonth"] = "1か月";
        this["FinancialChart_RangeSelector_OneYear"] = "1年";
        this["FinancialChart_RangeSelector_SixMonths"] = "6か月";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3か月";
        this["FinancialChart_RangeSelector_To"] = "終了:";
        this["FinancialChart_RangeSelector_YearToDate"] = "累計";
        this["FinancialSeries_Close"] = "終値";
        this["FinancialSeries_High"] = "高値";
        this["FinancialSeries_Low"] = "安値";
        this["FinancialSeries_Open"] = "始値";
        this["FinancialSeries_Volume"] = "ボリューム";
        this["NoEncodingsLoaded"] = "Infragistics.Encoding アセンブリを読み込むか、ヘルプのカスタム エンコーティングを作成する方法を参照してください。";
        this["NotSupportedEncoding"] = "{0} {1} に対するエンコード化はサポートされていません。";
        this["Object_Sealed"] = "オブジェクトがシールドされた後は変更できません。";
        this["OPD_DefaultInteraction"] = "マウス ドラッグ操作";
        this["OPD_DragPan"] = "サーフェスのパン";
        this["OPD_DragZoom"] = "選択領域にズーム";
        this["OPD_ScaleToFit"] = "画面に合わせる";
        this["OPD_ScaleToFit_SeriesViewer"] = "ズームのリセット (100%)";
        this["OPD_ZoomIn"] = "拡大";
        this["OPD_ZoomOut"] = "縮小";
        this["OPD_ZoomTo100"] = "100% にズーム";
        this["PieChart_Others"] = "その他";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "データ ソースの範囲を基にした変更は現在サポートされていません。";
        this["ScatterSeries_Value"] = "値";
        this["TRIAL_VERSION"] = "トライアル版 (購入は jp.infragistics.com)";
    }
}
