// igniteui-react-core/esm2015/lib/DataGridColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataGridColumnDescription = /*@__PURE__*/ (() => {
    class DataGridColumnDescription extends DefinitionBaseDescription {
        get_type() {
            return "DataGridColumn";
        }
        constructor() {
            super();
            this.d5 = 0;
            this.d7 = 0;
            this.d6 = 0;
            this.d4 = 0;
            this.e3 = null;
            this.ex = null;
            this.ez = null;
            this.el = null;
            this.e7 = null;
            this.ep = null;
            this.e5 = null;
            this.en = null;
            this.e6 = null;
            this.eo = null;
            this.cw = null;
            this.cs = null;
            this.dv = 0;
            this.da = false;
            this.c4 = false;
            this.cq = null;
            this.cu = null;
            this.cy = null;
            this.c9 = false;
            this.dc = false;
            this.db = false;
            this.dd = false;
            this.e8 = null;
            this.e4 = null;
            this.eh = null;
            this.eq = null;
            this.ej = null;
            this.es = null;
            this.er = null;
            this.ei = null;
            this.c1 = false;
            this.c5 = false;
            this.c3 = false;
            this.c7 = false;
            this.c2 = false;
            this.c6 = false;
            this.c8 = false;
            this.et = null;
            this.ek = null;
            this.du = 0;
            this.ds = 0;
            this.eu = null;
            this.dt = 0;
            this.ev = null;
            this.ew = null;
            this.e1 = null;
            this.e0 = null;
            this.e2 = null;
            this.d1 = 0;
            this.d3 = 0;
            this.d2 = 0;
            this.d0 = 0;
            this.ey = null;
            this.co = null;
            this.em = null;
        }
        get paddingLeft() {
            return this.d5;
        }
        set paddingLeft(a) {
            this.d5 = a;
            this.g("PaddingLeft");
        }
        get paddingTop() {
            return this.d7;
        }
        set paddingTop(a) {
            this.d7 = a;
            this.g("PaddingTop");
        }
        get paddingRight() {
            return this.d6;
        }
        set paddingRight(a) {
            this.d6 = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.d4;
        }
        set paddingBottom(a) {
            this.d4 = a;
            this.g("PaddingBottom");
        }
        get name() {
            return this.e3;
        }
        set name(a) {
            this.e3 = a;
            this.g("Name");
        }
        get field() {
            return this.ex;
        }
        set field(a) {
            this.ex = a;
            this.g("Field");
        }
        get headerText() {
            return this.ez;
        }
        set headerText(a) {
            this.ez = a;
            this.g("HeaderText");
        }
        get actualHeaderText() {
            return this.el;
        }
        set actualHeaderText(a) {
            this.el = a;
            this.g("ActualHeaderText");
        }
        get selectedBackground() {
            return this.e7;
        }
        set selectedBackground(a) {
            this.e7 = a;
            this.g("SelectedBackground");
        }
        get actualSelectedBackground() {
            return this.ep;
        }
        set actualSelectedBackground(a) {
            this.ep = a;
            this.g("ActualSelectedBackground");
        }
        get rowHoverBackground() {
            return this.e5;
        }
        set rowHoverBackground(a) {
            this.e5 = a;
            this.g("RowHoverBackground");
        }
        get actualHoverBackground() {
            return this.en;
        }
        set actualHoverBackground(a) {
            this.en = a;
            this.g("ActualHoverBackground");
        }
        get rowHoverTextColor() {
            return this.e6;
        }
        set rowHoverTextColor(a) {
            this.e6 = a;
            this.g("RowHoverTextColor");
        }
        get actualRowHoverTextColor() {
            return this.eo;
        }
        set actualRowHoverTextColor(a) {
            this.eo = a;
            this.g("ActualRowHoverTextColor");
        }
        get animationSettings() {
            return this.cw;
        }
        set animationSettings(a) {
            this.cw = a;
            this.g("AnimationSettings");
        }
        get width() {
            return this.cs;
        }
        set width(a) {
            this.cs = a;
            this.g("Width");
        }
        get minWidth() {
            return this.dv;
        }
        set minWidth(a) {
            this.dv = a;
            this.g("MinWidth");
        }
        get isFromMarkup() {
            return this.da;
        }
        set isFromMarkup(a) {
            this.da = a;
            this.g("IsFromMarkup");
        }
        get isAutoGenerated() {
            return this.c4;
        }
        set isAutoGenerated(a) {
            this.c4 = a;
            this.g("IsAutoGenerated");
        }
        get filter() {
            return this.cq;
        }
        set filter(a) {
            this.cq = a;
            this.g("Filter");
        }
        get filterExpression() {
            return this.cu;
        }
        set filterExpression(a) {
            this.cu = a;
            this.g("FilterExpression");
        }
        get header() {
            return this.cy;
        }
        set header(a) {
            this.cy = a;
            this.g("Header");
        }
        get isFilteringEnabled() {
            return this.c9;
        }
        set isFilteringEnabled(a) {
            this.c9 = a;
            this.g("IsFilteringEnabled");
        }
        get isResizingEnabled() {
            return this.dc;
        }
        set isResizingEnabled(a) {
            this.dc = a;
            this.g("IsResizingEnabled");
        }
        get isHidden() {
            return this.db;
        }
        set isHidden(a) {
            this.db = a;
            this.g("IsHidden");
        }
        get shouldRemoveWhenHidden() {
            return this.dd;
        }
        set shouldRemoveWhenHidden(a) {
            this.dd = a;
            this.g("ShouldRemoveWhenHidden");
        }
        get sortDirection() {
            return this.e8;
        }
        set sortDirection(a) {
            this.e8 = a;
            this.g("SortDirection");
        }
        get pinned() {
            return this.e4;
        }
        set pinned(a) {
            this.e4 = a;
            this.g("Pinned");
        }
        get actualColumnOptionsIconAlignment() {
            return this.eh;
        }
        set actualColumnOptionsIconAlignment(a) {
            this.eh = a;
            this.g("ActualColumnOptionsIconAlignment");
        }
        get columnOptionsIconAlignment() {
            return this.eq;
        }
        set columnOptionsIconAlignment(a) {
            this.eq = a;
            this.g("ColumnOptionsIconAlignment");
        }
        get actualColumnOptionsIconColor() {
            return this.ej;
        }
        set actualColumnOptionsIconColor(a) {
            this.ej = a;
            this.g("ActualColumnOptionsIconColor");
        }
        get columnOptionsIconColor() {
            return this.es;
        }
        set columnOptionsIconColor(a) {
            this.es = a;
            this.g("ColumnOptionsIconColor");
        }
        get columnOptionsIconBehavior() {
            return this.er;
        }
        set columnOptionsIconBehavior(a) {
            this.er = a;
            this.g("ColumnOptionsIconBehavior");
        }
        get actualColumnOptionsIconBehavior() {
            return this.ei;
        }
        set actualColumnOptionsIconBehavior(a) {
            this.ei = a;
            this.g("ActualColumnOptionsIconBehavior");
        }
        get actualIsColumnOptionsEnabled() {
            return this.c1;
        }
        set actualIsColumnOptionsEnabled(a) {
            this.c1 = a;
            this.g("ActualIsColumnOptionsEnabled");
        }
        get isColumnOptionsEnabled() {
            return this.c5;
        }
        set isColumnOptionsEnabled(a) {
            this.c5 = a;
            this.g("IsColumnOptionsEnabled");
        }
        get actualIsColumnOptionsSummariesEnabled() {
            return this.c3;
        }
        set actualIsColumnOptionsSummariesEnabled(a) {
            this.c3 = a;
            this.g("ActualIsColumnOptionsSummariesEnabled");
        }
        get isColumnOptionsSummariesEnabled() {
            return this.c7;
        }
        set isColumnOptionsSummariesEnabled(a) {
            this.c7 = a;
            this.g("IsColumnOptionsSummariesEnabled");
        }
        get actualIsColumnOptionsGroupingEnabled() {
            return this.c2;
        }
        set actualIsColumnOptionsGroupingEnabled(a) {
            this.c2 = a;
            this.g("ActualIsColumnOptionsGroupingEnabled");
        }
        get isColumnOptionsGroupingEnabled() {
            return this.c6;
        }
        set isColumnOptionsGroupingEnabled(a) {
            this.c6 = a;
            this.g("IsColumnOptionsGroupingEnabled");
        }
        get isEditable() {
            return this.c8;
        }
        set isEditable(a) {
            this.c8 = a;
            this.g("IsEditable");
        }
        get deletedTextColor() {
            return this.et;
        }
        set deletedTextColor(a) {
            this.et = a;
            this.g("DeletedTextColor");
        }
        get actualDeletedTextColor() {
            return this.ek;
        }
        set actualDeletedTextColor(a) {
            this.ek = a;
            this.g("ActualDeletedTextColor");
        }
        get editOpacity() {
            return this.du;
        }
        set editOpacity(a) {
            this.du = a;
            this.g("EditOpacity");
        }
        get actualEditOpacity() {
            return this.ds;
        }
        set actualEditOpacity(a) {
            this.ds = a;
            this.g("ActualEditOpacity");
        }
        get editFontFamily() {
            return this.eu;
        }
        set editFontFamily(a) {
            this.eu = a;
            this.g("EditFontFamily");
        }
        get editFontSize() {
            return this.dt;
        }
        set editFontSize(a) {
            this.dt = a;
            this.g("EditFontSize");
        }
        get editFontStyle() {
            return this.ev;
        }
        set editFontStyle(a) {
            this.ev = a;
            this.g("EditFontStyle");
        }
        get editFontWeight() {
            return this.ew;
        }
        set editFontWeight(a) {
            this.ew = a;
            this.g("EditFontWeight");
        }
        get mergedCellMode() {
            return this.e1;
        }
        set mergedCellMode(a) {
            this.e1 = a;
            this.g("MergedCellMode");
        }
        get mergedCellEvaluationCriteria() {
            return this.e0;
        }
        set mergedCellEvaluationCriteria(a) {
            this.e0 = a;
            this.g("MergedCellEvaluationCriteria");
        }
        get mergedCellVerticalAlignment() {
            return this.e2;
        }
        set mergedCellVerticalAlignment(a) {
            this.e2 = a;
            this.g("MergedCellVerticalAlignment");
        }
        get mergedCellPaddingLeft() {
            return this.d1;
        }
        set mergedCellPaddingLeft(a) {
            this.d1 = a;
            this.g("MergedCellPaddingLeft");
        }
        get mergedCellPaddingTop() {
            return this.d3;
        }
        set mergedCellPaddingTop(a) {
            this.d3 = a;
            this.g("MergedCellPaddingTop");
        }
        get mergedCellPaddingRight() {
            return this.d2;
        }
        set mergedCellPaddingRight(a) {
            this.d2 = a;
            this.g("MergedCellPaddingRight");
        }
        get mergedCellPaddingBottom() {
            return this.d0;
        }
        set mergedCellPaddingBottom(a) {
            this.d0 = a;
            this.g("MergedCellPaddingBottom");
        }
        get filterComparisonType() {
            return this.ey;
        }
        set filterComparisonType(a) {
            this.ey = a;
            this.g("FilterComparisonType");
        }
        get filterOperands() {
            return this.co;
        }
        set filterOperands(a) {
            this.co = a;
            this.g("FilterOperands");
        }
        get actualHeaderTextChangeRef() {
            return this.em;
        }
        set actualHeaderTextChangeRef(a) {
            this.em = a;
            this.g("ActualHeaderTextChangeRef");
        }
    }
    DataGridColumnDescription.$t = /*@__PURE__*/ markType(DataGridColumnDescription, 'DataGridColumnDescription', DefinitionBaseDescription.$);
    DataGridColumnDescription.__canMarshalByValue = true;
    DataGridColumnDescription.__marshalByValueAlias = "DataGridColumn";
    return DataGridColumnDescription;
})();
