// igniteui-react-core/esm2015/lib/DataGridCellEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataGridCellEventArgsDescription = /*@__PURE__*/ (() => {
    class DataGridCellEventArgsDescription extends Description {
        get_type() {
            return "DataGridCellEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = null;
            this.n = null;
            this.l = false;
        }
        get cellInfo() {
            return this.h;
        }
        set cellInfo(a) {
            this.h = a;
            this.g("CellInfo");
        }
        get grid() {
            return this.j;
        }
        set grid(a) {
            this.j = a;
            this.g("Grid");
        }
        get button() {
            return this.n;
        }
        set button(a) {
            this.n = a;
            this.g("Button");
        }
        get isDoubleClick() {
            return this.l;
        }
        set isDoubleClick(a) {
            this.l = a;
            this.g("IsDoubleClick");
        }
    }
    DataGridCellEventArgsDescription.$t = /*@__PURE__*/ markType(DataGridCellEventArgsDescription, 'DataGridCellEventArgsDescription', Description.$);
    return DataGridCellEventArgsDescription;
})();
