// igniteui-react-core/esm2015/lib/DataGridSummaryResultDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataGridSummaryResultDescription = /*@__PURE__*/ (() => {
    class DataGridSummaryResultDescription extends Description {
        get_type() {
            return "DataGridSummaryResult";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.p = null;
            this.m = null;
            this.k = 0;
            this.i = false;
        }
        get groupKeyRef() {
            return this.p;
        }
        set groupKeyRef(a) {
            this.p = a;
            this.g("GroupKeyRef");
        }
        get value() {
            return this.m;
        }
        set value(a) {
            this.m = a;
            this.g("Value");
        }
        get summaryIndex() {
            return this.k;
        }
        set summaryIndex(a) {
            this.k = a;
            this.g("SummaryIndex");
        }
        get shouldDisplay() {
            return this.i;
        }
        set shouldDisplay(a) {
            this.i = a;
            this.g("ShouldDisplay");
        }
    }
    DataGridSummaryResultDescription.$t = /*@__PURE__*/ markType(DataGridSummaryResultDescription, 'DataGridSummaryResultDescription', Description.$);
    DataGridSummaryResultDescription.__marshalByValue = true;
    DataGridSummaryResultDescription.__marshalByValueAlias = "DefaultSummaryResult";
    return DataGridSummaryResultDescription;
})();
