// igniteui-react-core/esm2015/lib/DataLegendSeriesValueInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { stringIsNullOrEmpty } from "./string";
/**
 * @hidden
 */
export let DataLegendSeriesValueInfo = /*@__PURE__*/ (() => {
    class DataLegendSeriesValueInfo extends Base {
        constructor() {
            super();
            this.l = 0;
            this.v = null;
            this.w = null;
            this.x = null;
            this.y = null;
            this.z = null;
            this.n = 0;
            this.m = 0;
            this.j = false;
            this.i = false;
            this.e = false;
            this.h = false;
            this.f = false;
            this.g = false;
            this.d = false;
            this.a = 0;
            this.o = 0;
            this.p = 0;
            this.k = false;
            this.c = false;
            this.b = false;
            this.r = null;
            this.q = null;
            this.s = null;
            this.t = null;
            this.b = true;
            this.c = true;
            this.k = false;
            this.v = "";
            this.w = "";
            this.x = "";
            this.y = "";
            this.s = "";
            this.n = -1;
            this.m = -1;
            this.j = false;
            this.i = false;
            this.e = true;
            this.h = true;
            this.f = true;
            this.g = true;
            this.d = true;
            this.z = null;
        }
        u() {
            let a = "";
            if (!stringIsNullOrEmpty(this.r)) {
                a += this.r + ": ";
            }
            else if (!stringIsNullOrEmpty(this.q)) {
                a += this.q + ": ";
            }
            a += this.l;
            return a;
        }
    }
    DataLegendSeriesValueInfo.$t = /*@__PURE__*/ markType(DataLegendSeriesValueInfo, 'DataLegendSeriesValueInfo');
    return DataLegendSeriesValueInfo;
})();
