// igniteui-react-core/esm2015/lib/DataSeriesAdapterProviderInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Delegate_$type, markType } from "./type";
import { List$1 } from "./List$1";
/**
 * @hidden
 */
export let DataSeriesAdapterProviderInfo = /*@__PURE__*/ (() => {
    class DataSeriesAdapterProviderInfo extends Base {
        constructor() {
            super();
            this.e = null;
            this.a = 0;
            this.c = null;
            this.b = null;
            this.d = false;
            this.e = new List$1(Delegate_$type, 0);
        }
    }
    DataSeriesAdapterProviderInfo.$t = /*@__PURE__*/ markType(DataSeriesAdapterProviderInfo, 'DataSeriesAdapterProviderInfo');
    return DataSeriesAdapterProviderInfo;
})();
