// igniteui-react-core/esm2015/lib/DataSourceSpecialRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataSourceSpecialRowDescription = /*@__PURE__*/ (() => {
    class DataSourceSpecialRowDescription extends Description {
        get_type() {
            return "DataSourceSpecialRow";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.p = null;
            this.m = null;
            this.k = 0;
            this.h = null;
        }
        get rowType() {
            return this.p;
        }
        set rowType(a) {
            this.p = a;
            this.g("RowType");
        }
        get targetRow() {
            return this.m;
        }
        set targetRow(a) {
            this.m = a;
            this.g("TargetRow");
        }
        get level() {
            return this.k;
        }
        set level(a) {
            this.k = a;
            this.g("Level");
        }
        get summaryResults() {
            return this.h;
        }
        set summaryResults(a) {
            this.h = a;
            this.g("SummaryResults");
        }
    }
    DataSourceSpecialRowDescription.$t = /*@__PURE__*/ markType(DataSourceSpecialRowDescription, 'DataSourceSpecialRowDescription', Description.$);
    DataSourceSpecialRowDescription.__marshalByValue = true;
    DataSourceSpecialRowDescription.__marshalByValueAlias = "DataSourceSpecialRow";
    return DataSourceSpecialRowDescription;
})();
