// igniteui-react-core/esm2015/lib/DataSourceRowExpansionChangedEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { EventArgs, markType } from "./type";
/**
 * @hidden
 */
export let DataSourceRowExpansionChangedEventArgs = /*@__PURE__*/ (() => {
    class DataSourceRowExpansionChangedEventArgs extends EventArgs {
        constructor(a, b, c) {
            super();
            this.f = -1;
            this.c = false;
            this.b = false;
            this.f = a;
            this.c = b;
            this.b = c;
        }
        get rowIndex() {
            return this.f;
        }
        get oldState() {
            return this.c;
        }
        get newState() {
            return this.b;
        }
    }
    DataSourceRowExpansionChangedEventArgs.$t = /*@__PURE__*/ markType(DataSourceRowExpansionChangedEventArgs, 'DataSourceRowExpansionChangedEventArgs', EventArgs.$);
    return DataSourceRowExpansionChangedEventArgs;
})();
