// igniteui-react-core/esm2015/lib/DataLegendSummaryColumn.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DataLegendSeriesValueInfo } from "./DataLegendSeriesValueInfo";
import { List$1 } from "./List$1";
import { HashSet$1 } from "./HashSet$1";
import { Number_$type, String_$type, markType } from "./type";
/**
 * @hidden
 */
export let DataLegendSummaryColumn = /*@__PURE__*/ (() => {
    class DataLegendSummaryColumn extends DataLegendSeriesValueInfo {
        constructor() {
            super();
            this.ah = new List$1(Number_$type, 0);
            this.af = new List$1(String_$type, 0);
            this.ag = new List$1(String_$type, 0);
            this.ad = new HashSet$1(String_$type, 0);
            this.ae = new HashSet$1(String_$type, 0);
            this.a = 15;
        }
        get ac() {
            return this.ah.toArray();
        }
        ak(a) {
            this.ah.add(a);
        }
        ai(a) {
            if (!this.ad.contains(a)) {
                this.ad.add_1(a);
                this.af.add(a);
            }
        }
        aj(a) {
            if (!this.ae.contains(a)) {
                this.ae.add_1(a);
                this.ag.add(a);
            }
        }
        get aa() {
            return this.af.toArray();
        }
        get ab() {
            return this.ag.toArray();
        }
    }
    DataLegendSummaryColumn.$t = /*@__PURE__*/ markType(DataLegendSummaryColumn, 'DataLegendSummaryColumn', DataLegendSeriesValueInfo.$);
    return DataLegendSummaryColumn;
})();
