// igniteui-react-core/esm2015/lib/DataSourceSummaryOperand.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
export var DataSourceSummaryOperand = /*@__PURE__*/ (function (DataSourceSummaryOperand) {
    DataSourceSummaryOperand[DataSourceSummaryOperand["Min"] = 0] = "Min";
    DataSourceSummaryOperand[DataSourceSummaryOperand["Max"] = 1] = "Max";
    DataSourceSummaryOperand[DataSourceSummaryOperand["Sum"] = 2] = "Sum";
    DataSourceSummaryOperand[DataSourceSummaryOperand["Average"] = 3] = "Average";
    DataSourceSummaryOperand[DataSourceSummaryOperand["Count"] = 4] = "Count";
    DataSourceSummaryOperand[DataSourceSummaryOperand["Custom"] = 5] = "Custom";
    return DataSourceSummaryOperand;
})({});
/**
 * @hidden
 */
export let DataSourceSummaryOperand_$type = /*@__PURE__*/ markEnum('DataSourceSummaryOperand', 'Min,0|Max,1|Sum,2|Average,3|Count,4|Custom,5');
