// igniteui-react-core/esm2015/lib/DataVisualizationLocaleFr.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
export class DataVisualizationLocaleFr {
    constructor() {
        this["AssemblyNotIncluded"] = "Impossible de charger {0}.dll. Vous devez vous assurer que cet ensemble est inclus dans le XAP qui est déployé sur le Windows Phone, par exemple, en lui ajoutant une référence depuis l'un de vos projets.";
        this["BreakEvenTitle_BreakEven"] = "Break Even";
        this["BreakEvenTitle_FixedCost"] = "Fixed Cost";
        this["BreakEvenTitle_LossArea"] = "Loss Area";
        this["BreakEvenTitle_MarginalProfit"] = "Marginal Profit";
        this["BreakEvenTitle_ProfitArea"] = "Profit Area";
        this["BreakEvenTitle_SafetyMargin"] = "Safety Margin";
        this["BreakEvenTitle_SalesRevenue"] = "Sales Revenue";
        this["BreakEvenTitle_TotalCost"] = "Total Cost";
        this["BreakEvenTitle_VariableCost"] = "Variable Cost";
        this["BubbleSeries_Radius"] = "Rayon";
        this["DataChart_InteractivityNotLoaded"] = "L’accès aux fonctionnalités d’interactivité a été possible mais le module d’interaction de graphiques n’a pas été chargé.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Les fonctionnalités d’abréviation numérique ont été utilisées, mais le module d’abréviation numérique n’a pas été chargé.";
        this["DataChart_VisualDataNotLoaded"] = "L’accès aux fonctionnalités de données visuelles a été possible mais le module d’interaction de graphiques n’a pas été chargé.";
        this["DataProviderNotInitialized"] = "Le fournisseur de données a été utilisé sans avoir été initialisé au préalable. Le fournisseur de données doit être initialisé avant toute utilisation.";
        this["DataSource_Summary_Avg"] = "Avg";
        this["DataSource_Summary_Count"] = "Count";
        this["DataSource_Summary_Max"] = "Max";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Sum";
        this["Default_Series_Title"] = "Titre de série";
        this["FinancialChart_IndicatorMenu_Header"] = "Indicateurs";
        this["FinancialChart_IndicatorMenu_Header1"] = "Indicateurs";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Indicateurs autonomes";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader1"] = "Indicateurs autonomes";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Superpositions";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader1"] = "Superpositions";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Lignes de tendance";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader1"] = "Lignes de tendance";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volume";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader1"] = "Volumes";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Oscillateur de volume absolu";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Accumulation-distribution";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Index directionnel moyen";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Average True Range";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Largeur de la bande de Bollinger";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Oscillateur de Chaikin";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Volatilité de Chaikin";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Commodity Channel Index";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Detrended Price Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Ease Of Movement";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Oscillateur stochastique rapide";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Force Index";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Oscillateur stochastique complet";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Market Facilitation Index";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Mass Index";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Prix médian";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Money Flow Index";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Negative Volume Index";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "Graphique du volume";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Oscillateur du pourcentage de prix";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Oscillateur du pourcentage de volume";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Positive Volume Index";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Price Volume Trend";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Oscillateur stochastique lent";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Écart type";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Prix typique";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Oscillateur ultime";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Clôture pondérée";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bandes de Bollinger";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Price Channel";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Ajustement cubique";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Moyenne cumulée";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Moyenne exponentielle";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Ajustement exponentiel";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Ajustement linéaire";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Ajustement logarithmique";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Moyenne modifiée";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "Ajustement de loi de puissance";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Ajustement quadratique";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Ajustement quartique";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Ajustement quintique";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Ajustement quintique";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Moyenne pondérée";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Zone";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Colonne";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Ligne";
        this["FinancialChart_RangeSelector_All"] = "Tous";
        this["FinancialChart_RangeSelector_From"] = "De";
        this["FinancialChart_RangeSelector_OneMonth"] = "1M";
        this["FinancialChart_RangeSelector_OneYear"] = "1A";
        this["FinancialChart_RangeSelector_SixMonths"] = "6M";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3M";
        this["FinancialChart_RangeSelector_To"] = "To";
        this["FinancialChart_RangeSelector_YearToDate"] = "AAJ";
        this["FinancialSeries_Close"] = "Fermer";
        this["FinancialSeries_High"] = "Haut";
        this["FinancialSeries_Low"] = "Bas";
        this["FinancialSeries_Open"] = "Ouvrir";
        this["FinancialSeries_Volume"] = "Volume";
        this["NoEncodingsLoaded"] = "Aucun encodage n'a été trouvé! Veuillez charger l'ensemble Infragistics.Encoding ou consultez la documentation concernant la création d'encodages personnalisés.";
        this["NotSupportedEncoding"] = "L'encodage correspondant sous la plage {0} {1} n'est pas pris en charge.";
        this["Object_Sealed"] = "Une fois que l’objet a été scellé, les modifications ne sont pas autorisées.";
        this["OPD_DefaultInteraction"] = "Sélectionner l'interaction de glissement de la souris par défaut";
        this["OPD_DragPan"] = "Faire un panoramique de la surface";
        this["OPD_DragZoom"] = "Faire un zoom avant sur la zone de sélection";
        this["OPD_ScaleToFit"] = "Adapter à l'écran";
        this["OPD_ScaleToFit_SeriesViewer"] = "Rétablir le zoom (à 100%)";
        this["OPD_ZoomIn"] = "Zoom avant";
        this["OPD_ZoomOut"] = "Zoom arrière";
        this["OPD_ZoomTo100"] = "Zoom à 100%";
        this["PieChart_Others"] = "Autres";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Les modifications basées sur la plage ne sont actuellement pas prises en charge pour la source des données";
        this["ScatterSeries_Value"] = "Valeur";
        this["TRIAL_VERSION"] = "Achat de la VERSION D’ESSAI SANS LICENCE sur Infragistics.com";
    }
}
