// igniteui-react-core/esm2015/lib/DataGridToolbarDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataGridToolbarDescription = /*@__PURE__*/ (() => {
    class DataGridToolbarDescription extends Description {
        get_type() {
            return "DataGridToolbar";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.ad = null;
            this.ae = null;
            this.ag = null;
            this.p = 0;
            this.ah = null;
            this.ai = null;
            this.af = null;
            this.v = null;
            this.ac = null;
            this.h = false;
            this.x = null;
            this.y = null;
            this.i = false;
            this.z = null;
            this.aa = null;
            this.l = 0;
            this.m = 0;
            this.n = 0;
            this.o = 0;
            this.ab = null;
            this.w = null;
        }
        get targetGridRef() {
            return this.ad;
        }
        set targetGridRef(a) {
            this.ad = a;
            this.g("TargetGridRef");
        }
        get toolbarTitle() {
            return this.ae;
        }
        set toolbarTitle(a) {
            this.ae = a;
            this.g("ToolbarTitle");
        }
        get toolbarTitleFontFamily() {
            return this.ag;
        }
        set toolbarTitleFontFamily(a) {
            this.ag = a;
            this.g("ToolbarTitleFontFamily");
        }
        get toolbarTitleFontSize() {
            return this.p;
        }
        set toolbarTitleFontSize(a) {
            this.p = a;
            this.g("ToolbarTitleFontSize");
        }
        get toolbarTitleFontStyle() {
            return this.ah;
        }
        set toolbarTitleFontStyle(a) {
            this.ah = a;
            this.g("ToolbarTitleFontStyle");
        }
        get toolbarTitleFontWeight() {
            return this.ai;
        }
        set toolbarTitleFontWeight(a) {
            this.ai = a;
            this.g("ToolbarTitleFontWeight");
        }
        get toolbarTitleColor() {
            return this.af;
        }
        set toolbarTitleColor(a) {
            this.af = a;
            this.g("ToolbarTitleColor");
        }
        get backgroundColor() {
            return this.v;
        }
        set backgroundColor(a) {
            this.v = a;
            this.g("BackgroundColor");
        }
        get dialogBackgroundColor() {
            return this.ac;
        }
        set dialogBackgroundColor(a) {
            this.ac = a;
            this.g("DialogBackgroundColor");
        }
        get columnChooser() {
            return this.h;
        }
        set columnChooser(a) {
            this.h = a;
            this.g("ColumnChooser");
        }
        get columnChooserText() {
            return this.x;
        }
        set columnChooserText(a) {
            this.x = a;
            this.g("ColumnChooserText");
        }
        get columnChooserTitle() {
            return this.y;
        }
        set columnChooserTitle(a) {
            this.y = a;
            this.g("ColumnChooserTitle");
        }
        get columnPinning() {
            return this.i;
        }
        set columnPinning(a) {
            this.i = a;
            this.g("ColumnPinning");
        }
        get columnPinningText() {
            return this.z;
        }
        set columnPinningText(a) {
            this.z = a;
            this.g("ColumnPinningText");
        }
        get columnPinningTitle() {
            return this.aa;
        }
        set columnPinningTitle(a) {
            this.aa = a;
            this.g("ColumnPinningTitle");
        }
        get borderWidthBottom() {
            return this.l;
        }
        set borderWidthBottom(a) {
            this.l = a;
            this.g("BorderWidthBottom");
        }
        get borderWidthLeft() {
            return this.m;
        }
        set borderWidthLeft(a) {
            this.m = a;
            this.g("BorderWidthLeft");
        }
        get borderWidthRight() {
            return this.n;
        }
        set borderWidthRight(a) {
            this.n = a;
            this.g("BorderWidthRight");
        }
        get borderWidthTop() {
            return this.o;
        }
        set borderWidthTop(a) {
            this.o = a;
            this.g("BorderWidthTop");
        }
        get density() {
            return this.ab;
        }
        set density(a) {
            this.ab = a;
            this.g("Density");
        }
        get baseTheme() {
            return this.w;
        }
        set baseTheme(a) {
            this.w = a;
            this.g("BaseTheme");
        }
    }
    DataGridToolbarDescription.$t = /*@__PURE__*/ markType(DataGridToolbarDescription, 'DataGridToolbarDescription', Description.$);
    return DataGridToolbarDescription;
})();
