// igniteui-react-core/esm2015/lib/DataVisualizationLocaleSv.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
export class DataVisualizationLocaleSv {
    constructor() {
        this["AssemblyNotIncluded"] = "Det går inte att ladda {0}.dll. Du måste se till att den här enheten ingår i XAP som distribueras till Windows Phone, till exempel genom att lägga till en referens till den från något av dina projekt.";
        this["BreakEvenTitle_BreakEven"] = "Gå jämt upp";
        this["BreakEvenTitle_FixedCost"] = "Fast kostnad";
        this["BreakEvenTitle_LossArea"] = "Förlustområde";
        this["BreakEvenTitle_MarginalProfit"] = "Marginal vinst";
        this["BreakEvenTitle_ProfitArea"] = "Vinstområde";
        this["BreakEvenTitle_SafetyMargin"] = "Säkerhetsmarginal";
        this["BreakEvenTitle_SalesRevenue"] = "Försäljningsintäkter";
        this["BreakEvenTitle_TotalCost"] = "Total kostnad";
        this["BreakEvenTitle_VariableCost"] = "Rörlig kostnad";
        this["BubbleSeries_Radius"] = "Radie";
        this["DataChart_InteractivityNotLoaded"] = "Interaktivitetsfunktionalitet nås men diagraminteraktionsmodulen har inte laddats.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Nummerförkortningsfunktionalitet användes, men nummerförkortningsmodulen har inte laddats.";
        this["DataChart_VisualDataNotLoaded"] = "VisualData-funktionalitet nås, men diagraminteraktionsmodulen har inte laddats.";
        this["DataProviderNotInitialized"] = "Dataleverantören användes utan att initialiseras först. Dataleverantören måste initieras före användning.";
        this["DataSource_Summary_Avg"] = "Genomsnitt";
        this["DataSource_Summary_Count"] = "Räkna";
        this["DataSource_Summary_Max"] = "Max";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Summa";
        this["Default_Series_Title"] = "Serietitel";
        this["FinancialChart_IndicatorMenu_Header"] = "Indikatorer";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Fristående indikatorer";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Överlägg";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Trendlinjer";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volymer";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Absolut volymoscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Ackumulering/Fördelning";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Genomsnittligt riktningsindex";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Genomsnittligt sant intervall";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Bollinger-bandvidd";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Chaikin-oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Chaikin-instabilitet";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Varukanalindex";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Avtrenderad prisoscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Rörelseenkelhet";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Snabb stokastisk oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Force-index";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Helstokastisk oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Marknadsförenklingsindex";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Massindex";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Medianpris";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Penningflödesindex";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Negativt volymindex";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "På balansvolym";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Procentuell oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Procent volymoscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Positivt volymindex";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Prisvolymstrend";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Långsam stokastisk oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Standardavvikelse";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Typiskt pris";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Ultimat oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Viktad stängning";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bollinger-band";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Priskanal";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Kubisk passform";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Kumulativt genomsnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Exponentiellt genomsnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Exponentiell passform";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Linjär passform";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Logaritmisk passform";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Modifierat genomsnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "PowerLaw-passform";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Kvadratisk passform";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Fjärdegradspassform";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Femtegradspassform";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Enkelt genomsnitt";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Vägt genomsnitt";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Område";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Kolumn";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Linje";
        this["FinancialChart_RangeSelector_All"] = "Allt";
        this["FinancialChart_RangeSelector_From"] = "Från";
        this["FinancialChart_RangeSelector_OneMonth"] = "1M";
        this["FinancialChart_RangeSelector_OneYear"] = "1Y";
        this["FinancialChart_RangeSelector_SixMonths"] = "6M";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3M";
        this["FinancialChart_RangeSelector_To"] = "Till";
        this["FinancialChart_RangeSelector_YearToDate"] = "YTD";
        this["FinancialSeries_Close"] = "Slut";
        this["FinancialSeries_High"] = "Hög";
        this["FinancialSeries_Low"] = "Låg";
        this["FinancialSeries_Open"] = "Öppna";
        this["FinancialSeries_Volume"] = "Volym";
        this["NoEncodingsLoaded"] = "Inga kodningar hittades! Ladda Infragistics.Encoding-monteringen eller titta i dokumentationen för att se hur man skapar anpassade kodningar.";
        this["NotSupportedEncoding"] = "Kodningen som motsvarar under {0} {1} stöds inte.";
        this["Object_Sealed"] = "Ändringar är inte tillåtna efter att objektet har förseglats.";
        this["OPD_DefaultInteraction"] = "Välj standard musinteraktion";
        this["OPD_DragPan"] = "Panorera ytan";
        this["OPD_DragZoom"] = "Zooma in i urvalsområdet";
        this["OPD_ScaleToFit"] = "Anpassa till skärm";
        this["OPD_ScaleToFit_SeriesViewer"] = "Återställ zoom (100%)";
        this["OPD_ZoomIn"] = "Zooma in";
        this["OPD_ZoomOut"] = "Zooma ut";
        this["OPD_ZoomTo100"] = "Zooma till 100%";
        this["PieChart_Others"] = "Andra";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Områdesbaserade modifieringar av datakällan stöds för närvarande inte";
        this["ScatterSeries_Value"] = "Värde";
        this["TRIAL_VERSION"] = "LICENSLÖS PROVNING Köp på Infragistics.com";
    }
}
