// igniteui-react-core/esm2015/lib/DataVisualizationLocalePt.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
export class DataVisualizationLocalePt {
    constructor() {
        this["AssemblyNotIncluded"] = "Não é possível carregar {0}.dll. É necessário garantir que esta assemblagem é incluída no XAP implementado no Windows Phone, adicionando-lhe, por exemplo, uma referência a partir de qualquer um dos seus projetos.";
        this["BreakEvenTitle_BreakEven"] = "Break Even";
        this["BreakEvenTitle_FixedCost"] = "Custo fixo";
        this["BreakEvenTitle_LossArea"] = "Área de perda";
        this["BreakEvenTitle_MarginalProfit"] = "Lucro marginal";
        this["BreakEvenTitle_ProfitArea"] = "Área de lucro";
        this["BreakEvenTitle_SafetyMargin"] = "Margem de segurança";
        this["BreakEvenTitle_SalesRevenue"] = "Receita de vendas";
        this["BreakEvenTitle_TotalCost"] = "Custo total";
        this["BreakEvenTitle_VariableCost"] = "Custo variável";
        this["BubbleSeries_Radius"] = "Raio";
        this["DataChart_InteractivityNotLoaded"] = "A funcionalidade de interatividade foi acedida, mas o módulo de interação do gráfico não foi carregado.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "A funcionalidade de abreviação de número foi usada, mas o módulo de abreviação de número não foi carregado.";
        this["DataChart_VisualDataNotLoaded"] = "A funcionalidade DadosVisuais foi acedida, mas o módulo de interação do gráfico não foi carregado.";
        this["DataProviderNotInitialized"] = "O fornecedor de dados foi utilizado sem inicializar primeiro. O fornecedor de dados deve ser inicializado antes da utilização.";
        this["DataSource_Summary_Avg"] = "Média";
        this["DataSource_Summary_Count"] = "Contagem";
        this["DataSource_Summary_Max"] = "Máx.";
        this["DataSource_Summary_Min"] = "Mín.";
        this["DataSource_Summary_Sum"] = "Soma";
        this["Default_Series_Title"] = "Título da série";
        this["FinancialChart_IndicatorMenu_Header"] = "Indicadores";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Indicadores independentes";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Sobreposições";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Linhas de tendência";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volumes";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Oscilador de volume absoluto";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Distribuição/Acumulação";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Índice direcional médio";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Intervalo médio verdadeiro";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Largura da banda de Bollinger";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Oscilador de Chaikin";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Volatilidade de Chaikin";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Índice do canal de mercadorias";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Oscilador de preço sem tendência";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Facilidade de movimento";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Oscilador estocástico rápido";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Índice de força";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Oscilador estocástico total";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Índice de facilitação do mercado";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Índice de massa";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Preço médio";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Índice de fluxo monetário";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Índice de volume negativo";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "Saldo de volume";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Oscilador de preço percentual";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Oscilador de volume percentual";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Índice de volume positivo";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Tendência de volume de preços";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Oscilador estocástico lento";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Desvio padrão";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Preço típico";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Oscilador final";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Fecho ponderado";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bandas de Bollinger";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Canal de preços";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Ajuste cúbico";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Média acumulada";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Média exponencial";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Ajuste exponencial";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Ajuste linear";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Ajuste logarítmico";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Média modificada";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "Ajuste de lei de potência";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Ajuste quadrático";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Ajuste de quarto grau";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Ajuste de quinto grau";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Média Simples";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Média ponderada";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Área";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Coluna";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Linha";
        this["FinancialChart_RangeSelector_All"] = "Todos";
        this["FinancialChart_RangeSelector_From"] = "De";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 mês";
        this["FinancialChart_RangeSelector_OneYear"] = "1 ano";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 meses";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 meses";
        this["FinancialChart_RangeSelector_To"] = "Para";
        this["FinancialChart_RangeSelector_YearToDate"] = "Ano até à data";
        this["FinancialSeries_Close"] = "Fechar";
        this["FinancialSeries_High"] = "Alto";
        this["FinancialSeries_Low"] = "Baixo";
        this["FinancialSeries_Open"] = "Aberto";
        this["FinancialSeries_Volume"] = "Volume";
        this["NoEncodingsLoaded"] = "Não foram encontradas codificações! Carregue a assemblagem Infragistics.Encoding ou consulte a documentação sobre como criar codificações personalizadas.";
        this["NotSupportedEncoding"] = "A codificação correspondente em {0} {1} não é suportada.";
        this["Object_Sealed"] = "Não são permitidas modificações após o fecho do objeto.";
        this["OPD_DefaultInteraction"] = "Selecionar a interação predefinida de arrasto do rato";
        this["OPD_DragPan"] = "Deslocar a superfície";
        this["OPD_DragZoom"] = "Ampliar para a área de seleção";
        this["OPD_ScaleToFit"] = "Ajustar ao ecrã";
        this["OPD_ScaleToFit_SeriesViewer"] = "Repor zoom (100%)";
        this["OPD_ZoomIn"] = "Ampliar";
        this["OPD_ZoomOut"] = "Reduzir";
        this["OPD_ZoomTo100"] = "Zoom para 100%";
        this["PieChart_Others"] = "Outros";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Neste momento, não há suporte para modificações com base no intervalo da fonte de dados";
        this["ScatterSeries_Value"] = "Valor";
        this["TRIAL_VERSION"] = "AVALIAÇÃO NÃO LICENCIADA Compre em Infragistics.com";
    }
}
