// igniteui-react-core/esm2015/lib/DataVisualizationLocaleKo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
export class DataVisualizationLocaleKo {
    constructor() {
        this["AssemblyNotIncluded"] = "{0}.dll 을 로딩할 수 없습니다. 예를 들면, 프로젝트에서 참조를 추가하여 이 어셈블리가 Windows Phone 에 배포된 XAP 에 포함되어 있는지 확인해야 합니다.";
        this["BreakEvenTitle_BreakEven"] = "손익분기점";
        this["BreakEvenTitle_FixedCost"] = "고정 비용";
        this["BreakEvenTitle_LossArea"] = "손실 영역";
        this["BreakEvenTitle_MarginalProfit"] = "한계 수익";
        this["BreakEvenTitle_ProfitArea"] = "수익 영역";
        this["BreakEvenTitle_SafetyMargin"] = "안전 여유";
        this["BreakEvenTitle_SalesRevenue"] = "매출 수입";
        this["BreakEvenTitle_TotalCost"] = "총 비용";
        this["BreakEvenTitle_VariableCost"] = "가변 비용";
        this["BubbleSeries_Radius"] = "범위";
        this["DataChart_InteractivityNotLoaded"] = "상호 작용 기능에 접속했지만 차트 상호 작용 모듈이 로딩되지 않았습니다.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "숫자 약어 기능을 사용했지만 숫자 약어 모듈이 로딩되지 않았습니다.";
        this["DataChart_VisualDataNotLoaded"] = "VisualData 기능에 접속했지만 차트 상호 작용 모듈이 로딩되지 않았습니다.";
        this["DataProviderNotInitialized"] = "데이터 제공자를 먼저 초기화하지 않고 사용했습니다. 사용하기 전에 데이터 제공자를 초기화해야 합니다.";
        this["DataSource_Summary_Avg"] = "평균";
        this["DataSource_Summary_Count"] = "총계";
        this["DataSource_Summary_Max"] = "최대";
        this["DataSource_Summary_Min"] = "최소";
        this["DataSource_Summary_Sum"] = "합계";
        this["Default_Series_Title"] = "시리즈 타이틀";
        this["FinancialChart_IndicatorMenu_Header"] = "인디케이터";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "독립형 인디케이터";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "오버레이";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "추세선";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "거래량";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "절대 거래량 지표";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "누적/분포";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "평균 지향 지수";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "평균 참 범위";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "볼린저 밴드 폭";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "차이킨 오실레이터";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "차이킨 변동성";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "커머더티 채널 인덱스";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "디트렌드 프라이스 오실레이터";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "이지 오브 무브먼트";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "패스트 스토캐스틱 오실레이터";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "강도 지수";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "풀 스토캐스틱 오실레이터";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "시장 촉진 지수";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "매스 인덱스";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "메디안 프라이스";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "자금 흐름 지표";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "네거티브 볼륨 인덱스";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "온 밸런스 볼륨";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "퍼센티지 프라이스 오실레이터";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "퍼센티지 볼륨 오실레이터";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "포지티브 볼륨 인덱스";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "프라이스 볼륨 트렌드";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "슬로 스토캐스틱 오실레이터";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "표준 편차";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "스토캐스틱 RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "티피컬 프라이스";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "얼티미트 오실레이터";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "웨이티드 클로즈";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "윌리엄스 %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "볼린저 밴드";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "프라이스 채널";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "큐빅 피팅";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "누적 평균";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "지수 평균";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "지수 피팅";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "선형 피팅";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "대수 피팅";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "수정된 평균";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "멱법칙 피팅";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "2 차 피팅";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "4 차 피팅";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "5 차 피팅";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "단순 평균";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "가중 평균";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "영역";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "열";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "선";
        this["FinancialChart_RangeSelector_All"] = "모두";
        this["FinancialChart_RangeSelector_From"] = "부터";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 개월";
        this["FinancialChart_RangeSelector_OneYear"] = "1 년";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 개월";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 개월";
        this["FinancialChart_RangeSelector_To"] = "에";
        this["FinancialChart_RangeSelector_YearToDate"] = "YTD (연간 누계)";
        this["FinancialSeries_Close"] = "닫기";
        this["FinancialSeries_High"] = "고";
        this["FinancialSeries_Low"] = "저";
        this["FinancialSeries_Open"] = "열기";
        this["FinancialSeries_Volume"] = "거래량";
        this["NoEncodingsLoaded"] = "인코딩을 찾을 수 없습니다! Infragistics 를 로딩해 주십시오. 어셈블리를 인코딩하거나 사용자 인코딩 작성 방법에 대한 설명서를 참조해 주십시오.";
        this["NotSupportedEncoding"] = "{0} {1} 에 해당하는 인코딩은 지원되지 않습니다.";
        this["Object_Sealed"] = "개체가 확정된 후에는 수정할 수 없습니다.";
        this["OPD_DefaultInteraction"] = "기본 마우스 드래그 상호 작용 선택";
        this["OPD_DragPan"] = "표면 패닝";
        this["OPD_DragZoom"] = "선택 영역 확대";
        this["OPD_ScaleToFit"] = "화면에 맞춤";
        this["OPD_ScaleToFit_SeriesViewer"] = "줌 재설정 (100%)";
        this["OPD_ZoomIn"] = "확대";
        this["OPD_ZoomOut"] = "축소";
        this["OPD_ZoomTo100"] = "줌 100%";
        this["PieChart_Others"] = "기타";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "데이터 소스의 범위를 기반으로 한 수정은 현재 지원되지 않습니다.";
        this["ScatterSeries_Value"] = "값";
        this["TRIAL_VERSION"] = "Infragistics.com 에서 라이선스가 없는 평가판 구입";
    }
}
