// igniteui-react-core/esm2015/lib/DataSeriesToDescriptionCustomization.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let DataSeriesToDescriptionCustomization = /*@__PURE__*/ (() => {
    class DataSeriesToDescriptionCustomization extends Base {
        constructor() {
            super();
            this._order = 0;
            this._matchType = null;
            this._propertyName = null;
            this._propertyValue = null;
            this._isCollectionInsertionAtEnd = false;
            this._isCollectionInsertionAtStart = false;
            this._isCollectionInsertionAtIndex = false;
            this._isCollectionRemovaltIndex = false;
            this._collectionIndex = 0;
            this._matchParentIndex = 0;
            this.e = null;
            this.order = -1;
            this.matchParentIndex = -1;
        }
        get order() {
            return this._order;
        }
        set order(a) {
            this._order = a;
        }
        get matchType() {
            return this._matchType;
        }
        set matchType(a) {
            this._matchType = a;
        }
        get propertyName() {
            return this._propertyName;
        }
        set propertyName(a) {
            this._propertyName = a;
        }
        get propertyValue() {
            return this._propertyValue;
        }
        set propertyValue(a) {
            this._propertyValue = a;
        }
        get isCollectionInsertionAtEnd() {
            return this._isCollectionInsertionAtEnd;
        }
        set isCollectionInsertionAtEnd(a) {
            this._isCollectionInsertionAtEnd = a;
        }
        get isCollectionInsertionAtStart() {
            return this._isCollectionInsertionAtStart;
        }
        set isCollectionInsertionAtStart(a) {
            this._isCollectionInsertionAtStart = a;
        }
        get isCollectionInsertionAtIndex() {
            return this._isCollectionInsertionAtIndex;
        }
        set isCollectionInsertionAtIndex(a) {
            this._isCollectionInsertionAtIndex = a;
        }
        get isCollectionRemovaltIndex() {
            return this._isCollectionRemovaltIndex;
        }
        set isCollectionRemovaltIndex(a) {
            this._isCollectionRemovaltIndex = a;
        }
        get collectionIndex() {
            return this._collectionIndex;
        }
        set collectionIndex(a) {
            this._collectionIndex = a;
        }
        get matchParentIndex() {
            return this._matchParentIndex;
        }
        set matchParentIndex(a) {
            this._matchParentIndex = a;
        }
    }
    DataSeriesToDescriptionCustomization.$t = /*@__PURE__*/ markType(DataSeriesToDescriptionCustomization, 'DataSeriesToDescriptionCustomization');
    return DataSeriesToDescriptionCustomization;
})();
