// igniteui-react-core/esm2015/lib/DataVisualizationLocaleZhHans.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
export class DataVisualizationLocaleZhHans {
    constructor() {
        this["AssemblyNotIncluded"] = "无法加载 {0}.dll。您需要确保该程序集包含在部署到 Windows Phone 的 XAP 中，例如，通过在您的任何项目中添加对该程序集的引用。";
        this["BreakEvenTitle_BreakEven"] = "收支平衡";
        this["BreakEvenTitle_FixedCost"] = "固定成本";
        this["BreakEvenTitle_LossArea"] = "损失区";
        this["BreakEvenTitle_MarginalProfit"] = "边际利润";
        this["BreakEvenTitle_ProfitArea"] = "利润区";
        this["BreakEvenTitle_SafetyMargin"] = "安全边际";
        this["BreakEvenTitle_SalesRevenue"] = "销售收入";
        this["BreakEvenTitle_TotalCost"] = "总成本";
        this["BreakEvenTitle_VariableCost"] = "可变成本";
        this["BubbleSeries_Radius"] = "半径";
        this["DataChart_InteractivityNotLoaded"] = "已访问交互功能，但尚未加载图表交互模块。";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "已使用数字缩写功能，但尚未加载数字缩写模块。";
        this["DataChart_VisualDataNotLoaded"] = "已访问 VisualData 功能，但尚未加载图表交互模块。";
        this["DataProviderNotInitialized"] = "使用数据提供者时未先初始化。数据提供者必须在使用前初始化。";
        this["DataSource_Summary_Avg"] = "平均";
        this["DataSource_Summary_Count"] = "计算";
        this["DataSource_Summary_Max"] = "最大";
        this["DataSource_Summary_Min"] = "最小";
        this["DataSource_Summary_Sum"] = "合计";
        this["Default_Series_Title"] = "系列标题";
        this["FinancialChart_IndicatorMenu_Header"] = "指标";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "独立指标";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "叠加";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "趋势线";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "成交量";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "绝对交易量震荡指标";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "累积分配";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "平均方向指数";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "平均真实范围";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "布林带宽度";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "佳庆摆动指标";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "佳庆波动性指标";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "商品渠道指数";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "去趋势价格震荡指标";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "轻松移动";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "快速随机摆动指标";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "力指数";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "全随机摆动指标";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "市场促进指数";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "质量指数";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "中位数价格";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "资金流量指数";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "平滑移动平均线 (MACD)";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "负成交量指数";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "平衡交易量指标";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "百分比价格震荡指标";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "百分比交易量震荡指标";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "正成交量指数";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "价量趋势";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "变动率 (ROC)";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "相对强弱指数 (RSI)";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "慢速随机摆动指标";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "标准偏差";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "随机相对强弱指数";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "三重指数平滑平均线 (TRIX)";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "典型价格";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "终极摆动指标";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "加权收盘价";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "威廉指标";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "布林带";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "价格通道";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "三次拟合";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "累积平均值";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "指数平均值";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "指数拟合";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "线性拟合";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "对数拟合";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "修正平均值";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "幂律拟合";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "二次拟合";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "四次拟合";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "五次拟合";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "简单平均值";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "加权平均值";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "面积图";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "列";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "线";
        this["FinancialChart_RangeSelector_All"] = "全部";
        this["FinancialChart_RangeSelector_From"] = "自";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 个月";
        this["FinancialChart_RangeSelector_OneYear"] = "1 年";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 个月";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 个月";
        this["FinancialChart_RangeSelector_To"] = "至";
        this["FinancialChart_RangeSelector_YearToDate"] = "年初至今";
        this["FinancialSeries_Close"] = "关闭";
        this["FinancialSeries_High"] = "盘高";
        this["FinancialSeries_Low"] = "盘点";
        this["FinancialSeries_Open"] = "打开";
        this["FinancialSeries_Volume"] = "成交量";
        this["NoEncodingsLoaded"] = "未找到编码！请加载 Infragistics.Encoding 程序集或查看文档中如何创建自定义编码。";
        this["NotSupportedEncoding"] = "不支持 {0} {1} 下的编码。";
        this["Object_Sealed"] = "密封物体后，不允许修改。";
        this["OPD_DefaultInteraction"] = "选择默认的鼠标拖动交互";
        this["OPD_DragPan"] = "平移表面";
        this["OPD_DragZoom"] = "缩放到所选区域";
        this["OPD_ScaleToFit"] = "适应屏幕";
        this["OPD_ScaleToFit_SeriesViewer"] = "重置缩放 (100%)";
        this["OPD_ZoomIn"] = "放大";
        this["OPD_ZoomOut"] = "缩小";
        this["OPD_ZoomTo100"] = "缩放至 100%";
        this["PieChart_Others"] = "其他";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "当前不支持基于范围的数据源修改";
        this["ScatterSeries_Value"] = "数值";
        this["TRIAL_VERSION"] = "未经许可的试用版，通过 Infragistics.com 购买";
    }
}
