// igniteui-react-core/esm2015/lib/DataVisualizationLocaleDa.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
export class DataVisualizationLocaleDa {
    constructor() {
        this["AssemblyNotIncluded"] = "Kan ikke indlæse {0} .dll. Du skal sørge for, at denne samling er inkluderet i XAP, der er implementeret på Windows Phone, for eksempel ved at tilføje en reference til den fra et af dine projekter.";
        this["BreakEvenTitle_BreakEven"] = "Omkostningsnulpunkt";
        this["BreakEvenTitle_FixedCost"] = "Faste omkostninger";
        this["BreakEvenTitle_LossArea"] = "Tabsområde";
        this["BreakEvenTitle_MarginalProfit"] = "Marginal fortjeneste";
        this["BreakEvenTitle_ProfitArea"] = "Profitområde";
        this["BreakEvenTitle_SafetyMargin"] = "Sikkerhedsmargen";
        this["BreakEvenTitle_SalesRevenue"] = "Salgsindtægter";
        this["BreakEvenTitle_TotalCost"] = "Udgifter i alt";
        this["BreakEvenTitle_VariableCost"] = "Variable omkostninger";
        this["BubbleSeries_Radius"] = "Radius";
        this["DataChart_InteractivityNotLoaded"] = "Der blev adgang til interaktivitetsfunktionalitet, men kortinteraktionsmodul er ikke indlæst.";
        this["DataChart_NumberAbbreviatorNotLoaded"] = "Nummerforkortelsesfunktionalitet blev brugt, men nummerforkortelsesmodulet er ikke indlæst.";
        this["DataChart_VisualDataNotLoaded"] = "VisualData-funktionalitet blev åbnet, men diagraminteraktionsmodulet er ikke indlæst.";
        this["DataProviderNotInitialized"] = "Dataudbyderen blev brugt uden først at initialisere. Dataudbyderen skal initialiseres inden brug.";
        this["DataSource_Summary_Avg"] = "Gns";
        this["DataSource_Summary_Count"] = "Tæl";
        this["DataSource_Summary_Max"] = "Maks";
        this["DataSource_Summary_Min"] = "Min";
        this["DataSource_Summary_Sum"] = "Sum";
        this["Default_Series_Title"] = "Serietitel";
        this["FinancialChart_IndicatorMenu_Header"] = "Indikatorer";
        this["FinancialChart_IndicatorMenu_IndicatorsCategoryHeader"] = "Frittstående indikatorer";
        this["FinancialChart_IndicatorMenu_OverlaysCategoryHeader"] = "Overlejringer";
        this["FinancialChart_IndicatorMenu_TrendlinesCategoryHeader"] = "Trendlinjer";
        this["FinancialChart_IndicatorMenu_VolumeCategoryHeader"] = "Volumener";
        this["FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator"] = "Absolute Volume Oscillator (volumenoscillator)";
        this["FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution"] = "Akkumulering/Fordeling";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex"] = "Gennemsnitligt retningsindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange"] = "Gennemsnitlig sand rækkevidde";
        this["FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth"] = "Bollinger-båndbredde";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator"] = "Chaikin-oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility"] = "Chaikin-volatilitet";
        this["FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex"] = "Varekanalindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator"] = "Detrended pris-oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement"] = "Let bevægelse";
        this["FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator"] = "Hurtig stokastisk oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_ForceIndex"] = "Kraft-indeks";
        this["FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator"] = "Fuld stokastisk oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex"] = "Indeks for markedsfacilitering";
        this["FinancialChart_IndicatorsMenu_Indicator_MassIndex"] = "Masseindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_MedianPrice"] = "Medianpris";
        this["FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex"] = "Pengeflow-indeks";
        this["FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence"] = "MACD";
        this["FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex"] = "Negativt volumenindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume"] = "På saldovolumen";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator"] = "Procentvis pris-oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator"] = "Procentvis volumenoscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex"] = "Positivt volumenindeks";
        this["FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend"] = "Prisvolumen Trend";
        this["FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum"] = "ROC";
        this["FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex"] = "RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator"] = "Langsom stokastisk oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_StandardDeviation"] = "Standardafvigelse";
        this["FinancialChart_IndicatorsMenu_Indicator_StochRSI"] = "Stoch RSI";
        this["FinancialChart_IndicatorsMenu_Indicator_TRIX"] = "TRIX";
        this["FinancialChart_IndicatorsMenu_Indicator_TypicalPrice"] = "Typisk pris";
        this["FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator"] = "Ultimate Oscillator";
        this["FinancialChart_IndicatorsMenu_Indicator_WeightedClose"] = "Vægtet luk";
        this["FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR"] = "Williams %R";
        this["FinancialChart_IndicatorsMenu_Overlay_BollingerBands"] = "Bollinger Bands";
        this["FinancialChart_IndicatorsMenu_Overlay_PriceChannel"] = "Priskanal";
        this["FinancialChart_IndicatorsMenu_TrendLine_CubicFit"] = "Kubisk tilpasning";
        this["FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage"] = "Kumulativt gennemsnit";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage"] = "Eksponentielt gennemsnit";
        this["FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit"] = "Eksponentiel tilpasning";
        this["FinancialChart_IndicatorsMenu_TrendLine_LinearFit"] = "Lineær tilpasning";
        this["FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit"] = "Logaritmisk tilpasning";
        this["FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage"] = "Modificeret gennemsnit";
        this["FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit"] = "PowerLaw tilpasning";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit"] = "Kvadratisk tilpasning";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuarticFit"] = "Kvartisk tilpasning";
        this["FinancialChart_IndicatorsMenu_TrendLine_QuinticFit"] = "Kvintisk tilpasning";
        this["FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage"] = "Simpel gennemsnit";
        this["FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage"] = "Vægtet gennemsnit";
        this["FinancialChart_IndicatorsMenu_Volume_Area"] = "Areal";
        this["FinancialChart_IndicatorsMenu_Volume_Column"] = "Kolonne";
        this["FinancialChart_IndicatorsMenu_Volume_Line"] = "Linie";
        this["FinancialChart_RangeSelector_All"] = "Alle";
        this["FinancialChart_RangeSelector_From"] = "Fra";
        this["FinancialChart_RangeSelector_OneMonth"] = "1 måned";
        this["FinancialChart_RangeSelector_OneYear"] = "1 år";
        this["FinancialChart_RangeSelector_SixMonths"] = "6 måneder";
        this["FinancialChart_RangeSelector_ThreeMonths"] = "3 måneder";
        this["FinancialChart_RangeSelector_To"] = "Til";
        this["FinancialChart_RangeSelector_YearToDate"] = "YTD (År-til-dato)";
        this["FinancialSeries_Close"] = "Tæt";
        this["FinancialSeries_High"] = "Høj";
        this["FinancialSeries_Low"] = "Lav";
        this["FinancialSeries_Open"] = "Åben";
        this["FinancialSeries_Volume"] = "Volumen";
        this["NoEncodingsLoaded"] = "Ingen kodninger fundet! Indlæs Infragistics.Encoding-samlingen eller se i dokumentationen, hvordan du opretter tilpassede kodninger.";
        this["NotSupportedEncoding"] = "Kodningen, der svarer til {0} {1}, understøttes ikke.";
        this["Object_Sealed"] = "Ændringer er ikke tilladt, efter at objektet er forseglet.";
        this["OPD_DefaultInteraction"] = "Vælg standard mus-træk-interaktion";
        this["OPD_DragPan"] = "Panorer overfladen";
        this["OPD_DragZoom"] = "Zoom ind i markeringsområdet";
        this["OPD_ScaleToFit"] = "Tilpas til skærm";
        this["OPD_ScaleToFit_SeriesViewer"] = "Nulstil zoom (100%)";
        this["OPD_ZoomIn"] = "Zoom ind";
        this["OPD_ZoomOut"] = "Zoom ud";
        this["OPD_ZoomTo100"] = "Zoom til 100%";
        this["PieChart_Others"] = "Andre";
        this["RangeModificationsNotSupportRangeModificationsNotSupported"] = "Intervalbaserede ændringer af datakilden understøttes i øjeblikket ikke";
        this["ScatterSeries_Value"] = "Værdi";
        this["TRIAL_VERSION"] = "ULICENSERET PRØVNING Køb på Infragistics.com";
    }
}
