// igniteui-react-core/esm2015/lib/DateTimeFormatSpecifierDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FormatSpecifierDescription } from "./FormatSpecifierDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let DateTimeFormatSpecifierDescription = /*@__PURE__*/ (() => {
    class DateTimeFormatSpecifierDescription extends FormatSpecifierDescription {
        get_type() {
            return "DateTimeFormatSpecifier";
        }
        constructor() {
            super();
            this.x = null;
            this.p = null;
            this.ad = null;
            this.l = 0;
            this.o = null;
            this.r = null;
            this.ab = null;
            this.y = null;
            this.ae = null;
            this.v = null;
            this.w = null;
            this.t = null;
            this.ag = null;
            this.s = null;
            this.ah = null;
            this.aa = null;
            this.q = null;
            this.u = null;
            this.z = null;
            this.ac = null;
            this.af = null;
        }
        get locale() {
            return this.x;
        }
        set locale(a) {
            this.x = a;
            this.g("Locale");
        }
        get dateStyle() {
            return this.p;
        }
        set dateStyle(a) {
            this.p = a;
            this.g("DateStyle");
        }
        get timeStyle() {
            return this.ad;
        }
        set timeStyle(a) {
            this.ad = a;
            this.g("TimeStyle");
        }
        get fractionalSecondDigits() {
            return this.l;
        }
        set fractionalSecondDigits(a) {
            this.l = a;
            this.g("FractionalSecondDigits");
        }
        get calendar() {
            return this.o;
        }
        set calendar(a) {
            this.o = a;
            this.g("Calendar");
        }
        get dayPeriod() {
            return this.r;
        }
        set dayPeriod(a) {
            this.r = a;
            this.g("DayPeriod");
        }
        get numberingSystem() {
            return this.ab;
        }
        set numberingSystem(a) {
            this.ab = a;
            this.g("NumberingSystem");
        }
        get localeMatcher() {
            return this.y;
        }
        set localeMatcher(a) {
            this.y = a;
            this.g("LocaleMatcher");
        }
        get timeZone() {
            return this.ae;
        }
        set timeZone(a) {
            this.ae = a;
            this.g("TimeZone");
        }
        get hour12() {
            return this.v;
        }
        set hour12(a) {
            this.v = a;
            this.g("Hour12");
        }
        get hourCycle() {
            return this.w;
        }
        set hourCycle(a) {
            this.w = a;
            this.g("HourCycle");
        }
        get formatMatcher() {
            return this.t;
        }
        set formatMatcher(a) {
            this.t = a;
            this.g("FormatMatcher");
        }
        get weekDay() {
            return this.ag;
        }
        set weekDay(a) {
            this.ag = a;
            this.g("WeekDay");
        }
        get era() {
            return this.s;
        }
        set era(a) {
            this.s = a;
            this.g("Era");
        }
        get year() {
            return this.ah;
        }
        set year(a) {
            this.ah = a;
            this.g("Year");
        }
        get month() {
            return this.aa;
        }
        set month(a) {
            this.aa = a;
            this.g("Month");
        }
        get day() {
            return this.q;
        }
        set day(a) {
            this.q = a;
            this.g("Day");
        }
        get hour() {
            return this.u;
        }
        set hour(a) {
            this.u = a;
            this.g("Hour");
        }
        get minute() {
            return this.z;
        }
        set minute(a) {
            this.z = a;
            this.g("Minute");
        }
        get second() {
            return this.ac;
        }
        set second(a) {
            this.ac = a;
            this.g("Second");
        }
        get timeZoneName() {
            return this.af;
        }
        set timeZoneName(a) {
            this.af = a;
            this.g("TimeZoneName");
        }
    }
    DateTimeFormatSpecifierDescription.$t = /*@__PURE__*/ markType(DateTimeFormatSpecifierDescription, 'DateTimeFormatSpecifierDescription', FormatSpecifierDescription.$);
    DateTimeFormatSpecifierDescription.__marshalByValue1 = true;
    DateTimeFormatSpecifierDescription.__marshalByValueAlias1 = "DateTimeFormatSpecifier";
    return DateTimeFormatSpecifierDescription;
})();
