// igniteui-react-core/esm2015/lib/DateRangeDescriptorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DateRangeDescriptorDescription = /*@__PURE__*/ (() => {
    class DateRangeDescriptorDescription extends Description {
        get_type() {
            return "DateRangeDescriptor";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.i = null;
        }
        get rangeType() {
            return this.l;
        }
        set rangeType(a) {
            this.l = a;
            this.g("RangeType");
        }
        get dateRange() {
            return this.i;
        }
        set dateRange(a) {
            this.i = a;
            this.g("DateRange");
        }
    }
    DateRangeDescriptorDescription.$t = /*@__PURE__*/ markType(DateRangeDescriptorDescription, 'DateRangeDescriptorDescription', Description.$);
    DateRangeDescriptorDescription.__marshalByValue = true;
    DateRangeDescriptorDescription.__marshalByValueAlias = "DateRangeDescriptor";
    return DateRangeDescriptorDescription;
})();
