// igniteui-react-core/esm2015/lib/DateTimeCellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CellInfoDescription } from "./CellInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let DateTimeCellInfoDescription = /*@__PURE__*/ (() => {
    class DateTimeCellInfoDescription extends CellInfoDescription {
        get_type() {
            return "DateTimeCellInfo";
        }
        constructor() {
            super();
            this.fq = new Date();
            this.fv = null;
            this.fl = null;
            this.fu = null;
            this.ft = null;
            this.fo = false;
        }
        get dateTimeValue() {
            return this.fq;
        }
        set dateTimeValue(a) {
            this.fq = a;
            this.g("DateTimeValue");
        }
        get formatStringOverride() {
            return this.fv;
        }
        set formatStringOverride(a) {
            this.fv = a;
            this.g("FormatStringOverride");
        }
        get formatSpecifiers() {
            return this.fl;
        }
        set formatSpecifiers(a) {
            this.fl = a;
            this.g("FormatSpecifiers");
        }
        get formatOverrideRef() {
            return this.fu;
        }
        set formatOverrideRef(a) {
            this.fu = a;
            this.g("FormatOverrideRef");
        }
        get dateTimeFormat() {
            return this.ft;
        }
        set dateTimeFormat(a) {
            this.ft = a;
            this.g("DateTimeFormat");
        }
        get isOffsetValue() {
            return this.fo;
        }
        set isOffsetValue(a) {
            this.fo = a;
            this.g("IsOffsetValue");
        }
    }
    DateTimeCellInfoDescription.$t = /*@__PURE__*/ markType(DateTimeCellInfoDescription, 'DateTimeCellInfoDescription', CellInfoDescription.$);
    DateTimeCellInfoDescription.__marshalByValue1 = true;
    DateTimeCellInfoDescription.__marshalByValueAlias1 = "DateTimeCellModel";
    return DateTimeCellInfoDescription;
})();
