// igniteui-react-core/esm2015/lib/DatePartDeltasDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DatePartDeltasDescription = /*@__PURE__*/ (() => {
    class DatePartDeltasDescription extends Description {
        get_type() {
            return "DatePartDeltas";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = 0;
            this.k = 0;
            this.m = 0;
            this.i = 0;
            this.j = 0;
            this.l = 0;
        }
        get date() {
            return this.h;
        }
        set date(a) {
            this.h = a;
            this.g("Date");
        }
        get month() {
            return this.k;
        }
        set month(a) {
            this.k = a;
            this.g("Month");
        }
        get year() {
            return this.m;
        }
        set year(a) {
            this.m = a;
            this.g("Year");
        }
        get hours() {
            return this.i;
        }
        set hours(a) {
            this.i = a;
            this.g("Hours");
        }
        get minutes() {
            return this.j;
        }
        set minutes(a) {
            this.j = a;
            this.g("Minutes");
        }
        get seconds() {
            return this.l;
        }
        set seconds(a) {
            this.l = a;
            this.g("Seconds");
        }
    }
    DatePartDeltasDescription.$t = /*@__PURE__*/ markType(DatePartDeltasDescription, 'DatePartDeltasDescription', Description.$);
    return DatePartDeltasDescription;
})();
