// igniteui-react-core/esm2015/lib/DateTimeColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DataGridColumnDescription } from "./DataGridColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let DateTimeColumnDescription = /*@__PURE__*/ (() => {
    class DateTimeColumnDescription extends DataGridColumnDescription {
        get_type() {
            return "DateTimeColumn";
        }
        constructor() {
            super();
            this.f9 = null;
            this.gg = null;
            this.gb = null;
            this.f1 = null;
            this.gf = null;
            this.f6 = false;
            this.gd = null;
            this.ga = null;
            this.f3 = null;
            this.gc = null;
            this.ge = null;
        }
        get dateTimeFormat() {
            return this.f9;
        }
        set dateTimeFormat(a) {
            this.f9 = a;
            this.g("DateTimeFormat");
        }
        get formatString() {
            return this.gg;
        }
        set formatString(a) {
            this.gg = a;
            this.g("FormatString");
        }
        get editorFormatString() {
            return this.gb;
        }
        set editorFormatString(a) {
            this.gb = a;
            this.g("EditorFormatString");
        }
        get formatSpecifiers() {
            return this.f1;
        }
        set formatSpecifiers(a) {
            this.f1 = a;
            this.g("FormatSpecifiers");
        }
        get formatOverrideRef() {
            return this.gf;
        }
        set formatOverrideRef(a) {
            this.gf = a;
            this.g("FormatOverrideRef");
        }
        get showTodayButton() {
            return this.f6;
        }
        set showTodayButton(a) {
            this.f6 = a;
            this.g("ShowTodayButton");
        }
        get editorType() {
            return this.gd;
        }
        set editorType(a) {
            this.gd = a;
            this.g("EditorType");
        }
        get editorDataSourceRef() {
            return this.ga;
        }
        set editorDataSourceRef(a) {
            this.ga = a;
            this.g("EditorDataSourceRef");
        }
        get editorDataSource() {
            return this.f3;
        }
        set editorDataSource(a) {
            this.f3 = a;
            this.g("EditorDataSource");
        }
        get editorTextField() {
            return this.gc;
        }
        set editorTextField(a) {
            this.gc = a;
            this.g("EditorTextField");
        }
        get editorValueField() {
            return this.ge;
        }
        set editorValueField(a) {
            this.ge = a;
            this.g("EditorValueField");
        }
    }
    DateTimeColumnDescription.$t = /*@__PURE__*/ markType(DateTimeColumnDescription, 'DateTimeColumnDescription', DataGridColumnDescription.$);
    DateTimeColumnDescription.__canMarshalByValue1 = true;
    DateTimeColumnDescription.__marshalByValueAlias1 = "DateTimeColumn";
    return DateTimeColumnDescription;
})();
