// igniteui-react-core/esm2015/lib/Decoder.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { NotImplementedException } from "./NotImplementedException";
/**
 * @hidden
 */
export let Decoder = /*@__PURE__*/ (() => {
    class Decoder extends Base {
        e(a, b, c, d, e, f, g, h, i, j) {
            throw new NotImplementedException(0);
            return {
                p7: h,
                p8: i,
                p9: j
            };
        }
        b(a, b, c, d) {
            throw new NotImplementedException(0);
        }
        d(a, b, c, d, e, f) {
            throw new NotImplementedException(0);
        }
        f() {
            throw new NotImplementedException(0);
        }
    }
    Decoder.$t = /*@__PURE__*/ markType(Decoder, 'Decoder');
    return Decoder;
})();
