// igniteui-react-core/esm2015/lib/DefaultSectionInformation.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { ISectionInformation_$type } from "./ISectionInformation";
/**
 * @hidden
 */
export let DefaultSectionInformation = /*@__PURE__*/ (() => {
    class DefaultSectionInformation extends Base {
        constructor(a, b, c, d, e) {
            super();
            this.e = 0;
            this.d = 0;
            this.a = null;
            this.b = null;
            this.c = null;
            this.e = a;
            this.d = b;
            this.a = c;
            this.b = d;
            this.c = e;
        }
        get startIndex() {
            return this.e;
        }
        get endIndex() {
            return this.d;
        }
        get groupKeyValues() {
            return this.b;
        }
        get groupKeyProperties() {
            return this.a;
        }
        get summaryResults() {
            return this.c;
        }
    }
    DefaultSectionInformation.$t = /*@__PURE__*/ markType(DefaultSectionInformation, 'DefaultSectionInformation', Base.$, [ISectionInformation_$type]);
    return DefaultSectionInformation;
})();
