// igniteui-react-core/esm2015/lib/DefaultSummaryResult.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { ISummaryResult_$type } from "./ISummaryResult";
/**
 * @hidden
 */
export let DefaultSummaryResult = /*@__PURE__*/ (() => {
    class DefaultSummaryResult extends Base {
        constructor(a, b, c) {
            super();
            this.a = null;
            this.g = null;
            this.f = null;
            this.b = 0;
            this.e = 0;
            this.d = false;
            this.d = true;
            this.g = a;
            this.b = b;
            this.f = c;
        }
        static createWithDisplay(a, b, c, d) {
            let e = new DefaultSummaryResult(a, b, c);
            e.shouldDisplay = d;
            return e;
        }
        get groupKey() {
            return this.a;
        }
        set groupKey(a) {
            this.a = a;
        }
        get propertyName() {
            return this.g;
        }
        get operand() {
            return this.b;
        }
        get value() {
            return this.f;
        }
        set value(a) {
            this.f = a;
        }
        get summaryIndex() {
            return this.e;
        }
        set summaryIndex(a) {
            this.e = a;
        }
        get shouldDisplay() {
            return this.d;
        }
        set shouldDisplay(a) {
            this.d = a;
        }
    }
    DefaultSummaryResult.$t = /*@__PURE__*/ markType(DefaultSummaryResult, 'DefaultSummaryResult', Base.$, [ISummaryResult_$type]);
    return DefaultSummaryResult;
})();
