// igniteui-react-core/esm2015/lib/DefinitionBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DefinitionBaseDescription = /*@__PURE__*/ (() => {
    class DefinitionBaseDescription extends Description {
        get_type() {
            return "DefinitionBase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.bb = null;
            this.h = null;
            this.be = null;
            this.az = null;
            this.bi = null;
            this.a4 = null;
            this.ac = 0;
            this.ae = 0;
            this.ad = 0;
            this.ab = 0;
            this.u = 0;
            this.w = 0;
            this.v = 0;
            this.t = 0;
            this.ag = 0;
            this.ai = 0;
            this.ah = 0;
            this.af = 0;
            this.a1 = null;
            this.a0 = null;
            this.a5 = null;
            this.bq = null;
            this.a9 = null;
            this.bp = null;
            this.a8 = null;
            this.bn = null;
            this.a6 = null;
            this.m = 0;
            this.bm = null;
            this.bs = null;
            this.bo = null;
            this.a7 = null;
            this.br = null;
            this.ba = null;
            this.j = false;
            this.bc = null;
            this.a2 = null;
            this.bd = null;
            this.a3 = null;
            this.aa = 0;
            this.y = 0;
            this.z = 0;
            this.x = 0;
            this.o = 0;
            this.l = 0;
            this.bj = null;
            this.n = 0;
            this.bk = null;
            this.bl = null;
            this.bg = null;
            this.bh = null;
            this.bf = null;
        }
        get background() {
            return this.bb;
        }
        set background(a) {
            this.bb = a;
            this.g("Background");
        }
        get conditionalStyles() {
            return this.h;
        }
        set conditionalStyles(a) {
            this.h = a;
            this.g("ConditionalStyles");
        }
        get border() {
            return this.be;
        }
        set border(a) {
            this.be = a;
            this.g("Border");
        }
        get activationBorder() {
            return this.az;
        }
        set activationBorder(a) {
            this.az = a;
            this.g("ActivationBorder");
        }
        get errorBorder() {
            return this.bi;
        }
        set errorBorder(a) {
            this.bi = a;
            this.g("ErrorBorder");
        }
        get actualBorder() {
            return this.a4;
        }
        set actualBorder(a) {
            this.a4 = a;
            this.g("ActualBorder");
        }
        get borderLeftWidth() {
            return this.ac;
        }
        set borderLeftWidth(a) {
            this.ac = a;
            this.g("BorderLeftWidth");
        }
        get borderTopWidth() {
            return this.ae;
        }
        set borderTopWidth(a) {
            this.ae = a;
            this.g("BorderTopWidth");
        }
        get borderRightWidth() {
            return this.ad;
        }
        set borderRightWidth(a) {
            this.ad = a;
            this.g("BorderRightWidth");
        }
        get borderBottomWidth() {
            return this.ab;
        }
        set borderBottomWidth(a) {
            this.ab = a;
            this.g("BorderBottomWidth");
        }
        get activationBorderLeftWidth() {
            return this.u;
        }
        set activationBorderLeftWidth(a) {
            this.u = a;
            this.g("ActivationBorderLeftWidth");
        }
        get activationBorderTopWidth() {
            return this.w;
        }
        set activationBorderTopWidth(a) {
            this.w = a;
            this.g("ActivationBorderTopWidth");
        }
        get activationBorderRightWidth() {
            return this.v;
        }
        set activationBorderRightWidth(a) {
            this.v = a;
            this.g("ActivationBorderRightWidth");
        }
        get activationBorderBottomWidth() {
            return this.t;
        }
        set activationBorderBottomWidth(a) {
            this.t = a;
            this.g("ActivationBorderBottomWidth");
        }
        get errorBorderLeftWidth() {
            return this.ag;
        }
        set errorBorderLeftWidth(a) {
            this.ag = a;
            this.g("ErrorBorderLeftWidth");
        }
        get errorBorderTopWidth() {
            return this.ai;
        }
        set errorBorderTopWidth(a) {
            this.ai = a;
            this.g("ErrorBorderTopWidth");
        }
        get errorBorderRightWidth() {
            return this.ah;
        }
        set errorBorderRightWidth(a) {
            this.ah = a;
            this.g("ErrorBorderRightWidth");
        }
        get errorBorderBottomWidth() {
            return this.af;
        }
        set errorBorderBottomWidth(a) {
            this.af = a;
            this.g("ErrorBorderBottomWidth");
        }
        get actualBackground() {
            return this.a1;
        }
        set actualBackground(a) {
            this.a1 = a;
            this.g("ActualBackground");
        }
        get actualActivationBorder() {
            return this.a0;
        }
        set actualActivationBorder(a) {
            this.a0 = a;
            this.g("ActualActivationBorder");
        }
        get actualErrorBorder() {
            return this.a5;
        }
        set actualErrorBorder(a) {
            this.a5 = a;
            this.g("ActualErrorBorder");
        }
        get stickyRowBackground() {
            return this.bq;
        }
        set stickyRowBackground(a) {
            this.bq = a;
            this.g("StickyRowBackground");
        }
        get actualStickyRowBackground() {
            return this.a9;
        }
        set actualStickyRowBackground(a) {
            this.a9 = a;
            this.g("ActualStickyRowBackground");
        }
        get pinnedRowBackground() {
            return this.bp;
        }
        set pinnedRowBackground(a) {
            this.bp = a;
            this.g("PinnedRowBackground");
        }
        get actualPinnedRowBackground() {
            return this.a8;
        }
        set actualPinnedRowBackground(a) {
            this.a8 = a;
            this.g("ActualPinnedRowBackground");
        }
        get lastStickyRowBackground() {
            return this.bn;
        }
        set lastStickyRowBackground(a) {
            this.bn = a;
            this.g("LastStickyRowBackground");
        }
        get actualLastStickyRowBackground() {
            return this.a6;
        }
        set actualLastStickyRowBackground(a) {
            this.a6 = a;
            this.g("ActualLastStickyRowBackground");
        }
        get contentOpacity() {
            return this.m;
        }
        set contentOpacity(a) {
            this.m = a;
            this.g("ContentOpacity");
        }
        get horizontalAlignment() {
            return this.bm;
        }
        set horizontalAlignment(a) {
            this.bm = a;
            this.g("HorizontalAlignment");
        }
        get verticalAlignment() {
            return this.bs;
        }
        set verticalAlignment(a) {
            this.bs = a;
            this.g("VerticalAlignment");
        }
        get lineBreakMode() {
            return this.bo;
        }
        set lineBreakMode(a) {
            this.bo = a;
            this.g("LineBreakMode");
        }
        get actualLineBreakMode() {
            return this.a7;
        }
        set actualLineBreakMode(a) {
            this.a7 = a;
            this.g("ActualLineBreakMode");
        }
        get textColor() {
            return this.br;
        }
        set textColor(a) {
            this.br = a;
            this.g("TextColor");
        }
        get actualTextColor() {
            return this.ba;
        }
        set actualTextColor(a) {
            this.ba = a;
            this.g("ActualTextColor");
        }
        get isBarSupported() {
            return this.j;
        }
        set isBarSupported(a) {
            this.j = a;
            this.g("IsBarSupported");
        }
        get barBackground() {
            return this.bc;
        }
        set barBackground(a) {
            this.bc = a;
            this.g("BarBackground");
        }
        get actualBarBackground() {
            return this.a2;
        }
        set actualBarBackground(a) {
            this.a2 = a;
            this.g("ActualBarBackground");
        }
        get barOutline() {
            return this.bd;
        }
        set barOutline(a) {
            this.bd = a;
            this.g("BarOutline");
        }
        get actualBarOutline() {
            return this.a3;
        }
        set actualBarOutline(a) {
            this.a3 = a;
            this.g("ActualBarOutline");
        }
        get barStrokeThickness() {
            return this.aa;
        }
        set barStrokeThickness(a) {
            this.aa = a;
            this.g("BarStrokeThickness");
        }
        get actualBarStrokeThickness() {
            return this.y;
        }
        set actualBarStrokeThickness(a) {
            this.y = a;
            this.g("ActualBarStrokeThickness");
        }
        get barCornerRadius() {
            return this.z;
        }
        set barCornerRadius(a) {
            this.z = a;
            this.g("BarCornerRadius");
        }
        get actualBarCornerRadius() {
            return this.x;
        }
        set actualBarCornerRadius(a) {
            this.x = a;
            this.g("ActualBarCornerRadius");
        }
        get pinnedRowOpacity() {
            return this.o;
        }
        set pinnedRowOpacity(a) {
            this.o = a;
            this.g("PinnedRowOpacity");
        }
        get actualPinnedRowOpacity() {
            return this.l;
        }
        set actualPinnedRowOpacity(a) {
            this.l = a;
            this.g("ActualPinnedRowOpacity");
        }
        get fontFamily() {
            return this.bj;
        }
        set fontFamily(a) {
            this.bj = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.n;
        }
        set fontSize(a) {
            this.n = a;
            this.g("FontSize");
        }
        get fontStyle() {
            return this.bk;
        }
        set fontStyle(a) {
            this.bk = a;
            this.g("FontStyle");
        }
        get fontWeight() {
            return this.bl;
        }
        set fontWeight(a) {
            this.bl = a;
            this.g("FontWeight");
        }
        get dataBindingRef() {
            return this.bg;
        }
        set dataBindingRef(a) {
            this.bg = a;
            this.g("DataBindingRef");
        }
        get dataBoundRef() {
            return this.bh;
        }
        set dataBoundRef(a) {
            this.bh = a;
            this.g("DataBoundRef");
        }
        get cellStyleKeyRequestedRef() {
            return this.bf;
        }
        set cellStyleKeyRequestedRef(a) {
            this.bf = a;
            this.g("CellStyleKeyRequestedRef");
        }
    }
    DefinitionBaseDescription.$t = /*@__PURE__*/ markType(DefinitionBaseDescription, 'DefinitionBaseDescription', Description.$);
    return DefinitionBaseDescription;
})();
