// igniteui-react-core/esm2015/lib/Description.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, String_$type, Boolean_$type, markType } from "./type";
import { List$1 } from "./List$1";
import { Dictionary$2 } from "./Dictionary$2";
/**
 * @hidden
 */
export let Description = /*@__PURE__*/ (() => {
    class Description extends Base {
        constructor() {
            super(...arguments);
            this.c = new Dictionary$2(String_$type, Boolean_$type, 0);
            this.d = new List$1(String_$type, 0);
            this.e = null;
        }
        a() {
            if (this.d.count == 0) {
                this.d.add("Type");
            }
            return this.d.toArray();
        }
        g(a) {
            if (!this.c.containsKey(a)) {
                if (this.d.count == 0) {
                    this.d.add("Type");
                }
                this.d.add(a);
            }
            this.c.item(a, true);
        }
        b(a) {
            if (this.c.containsKey(a)) {
                return this.c.item(a);
            }
            return false;
        }
        get name() {
            return this.e;
        }
        set name(a) {
            this.e = a;
            this.g("Name");
        }
    }
    Description.$t = /*@__PURE__*/ markType(Description, 'Description');
    return Description;
})();
