// igniteui-react-core/esm2015/lib/DescriptionTreeReconciler_TreeModeData.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let DescriptionTreeReconciler_TreeModeData = /*@__PURE__*/ (() => {
    class DescriptionTreeReconciler_TreeModeData extends Base {
        constructor() {
            super(...arguments);
            this.f = null;
            this.a = null;
            this.d = 0;
            this.c = false;
            this.e = 0;
            this.b = false;
        }
    }
    DescriptionTreeReconciler_TreeModeData.$t = /*@__PURE__*/ markType(DescriptionTreeReconciler_TreeModeData, 'DescriptionTreeReconciler_TreeModeData');
    return DescriptionTreeReconciler_TreeModeData;
})();
