// igniteui-react-core/esm2015/lib/DescriptionTreeAction.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { TypeDescriptionMetadata } from "./TypeDescriptionMetadata";
import { stringEndsWith } from "./string";
/**
 * @hidden
 */
export let DescriptionTreeAction = /*@__PURE__*/ (() => {
    class DescriptionTreeAction extends Base {
        constructor() {
            super(...arguments);
            this.a = 0;
            this.c = null;
            this.b = null;
            this.d = null;
            this.j = null;
            this.h = null;
            this.g = null;
            this.f = 0;
            this.e = 0;
        }
        i(a) {
            let b = this.j;
            if (this.d != null) {
                b = this.d.k(a);
            }
            else {
                if (TypeDescriptionMetadata.shouldCamelize(a)) {
                    b = TypeDescriptionMetadata.camelize(this.j);
                }
            }
            if (stringEndsWith(b, "Ref")) {
                b = b.substr(0, b.length - ("Ref").length);
            }
            return b;
        }
    }
    DescriptionTreeAction.$t = /*@__PURE__*/ markType(DescriptionTreeAction, 'DescriptionTreeAction');
    return DescriptionTreeAction;
})();
