// igniteui-react-core/esm2015/lib/DescriptionRef.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { List$1 } from "./List$1";
import { RefValueChangedTarget } from "./RefValueChangedTarget";
import { DescriptionRefValueChangedEventArgs } from "./DescriptionRefValueChangedEventArgs";
/**
 * @hidden
 */
export let DescriptionRef = /*@__PURE__*/ (() => {
    class DescriptionRef extends Base {
        constructor(a) {
            super();
            this.d = null;
            this.b = null;
            this.a = false;
            this.d = a;
            this.b = new List$1(RefValueChangedTarget.$, 0);
        }
        get e() {
            return this.d;
        }
        get c() {
            return this.b.count;
        }
        g(a, b, c, d) {
            this.b.add(((() => {
                let $ret = new RefValueChangedTarget();
                $ret.b = a;
                $ret.c = b;
                $ret.d = c;
                $ret.a = d;
                return $ret;
            })()));
        }
        j(a, b, c) {
            for (let d = this.b.count - 1; d >= 0; d--) {
                if (this.b._inner[d].c == b && this.b._inner[d].d == c) {
                    this.b.removeAt(d);
                }
            }
        }
        l(a) {
            for (let b = this.b.count - 1; b >= 0; b--) {
                if (this.b._inner[b].c == a) {
                    this.b.removeAt(b);
                }
            }
        }
        k(a) {
            for (let b = this.b.count - 1; b >= 0; b--) {
                if (this.b._inner[b].b == a) {
                    this.b.removeAt(b);
                }
            }
        }
        h() {
            this.b.clear();
        }
        i(a) {
            for (let b = this.b.count - 1; b >= 0; b--) {
                if (this.b._inner[b].c == a) {
                    this.b.removeAt(b);
                }
            }
        }
        f(a, b) {
            for (let c = 0; c < this.b.count; c++) {
                this.b._inner[c].a(this, ((() => {
                    let $ret = new DescriptionRefValueChangedEventArgs();
                    $ret.b = a;
                    $ret.a = b;
                    return $ret;
                })()));
            }
        }
    }
    DescriptionRef.$t = /*@__PURE__*/ markType(DescriptionRef, 'DescriptionRef');
    return DescriptionRef;
})();
