// igniteui-react-core/esm2015/lib/DivBridge.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIElementBridge } from "./NativeUIElementBridge";
import { INativeUIContentBridge_$type } from "./INativeUIContentBridge";
import { markType } from "./type";
/**
 * @hidden
 */
export let DivBridge = /*@__PURE__*/ (() => {
    class DivBridge extends NativeUIElementBridge {
        constructor() {
            super(...arguments);
            this.s = null;
        }
        getContent(a) {
            return this.s;
        }
        getValue(a, b) {
            return null;
        }
        setContent(a, b) {
            if (this.s == b) {
                return;
            }
            this.s = b;
            if (b == null) {
                for (let c = a.getChildCount() - 1; c >= 0; c--) {
                    let d = a.getChildAt(c);
                    d.remove();
                }
            }
            else {
                for (let e = a.getChildCount() - 1; e >= 0; e--) {
                    let f = a.getChildAt(e);
                    f.remove();
                }
                a.append(b.an);
            }
        }
        setValue(a, b, c) {
        }
    }
    DivBridge.$t = /*@__PURE__*/ markType(DivBridge, 'DivBridge', NativeUIElementBridge.$, [INativeUIContentBridge_$type]);
    return DivBridge;
})();
