// igniteui-react-core/esm2015/lib/DomainChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DomainChartDescription = /*@__PURE__*/ (() => {
    class DomainChartDescription extends Description {
        get_type() {
            return "DomainChart";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.fg = 0;
            this.ft = 0;
            this.fu = 0;
            this.fv = 0;
            this.fs = 0;
            this.fo = 0;
            this.fq = 0;
            this.fp = 0;
            this.fn = 0;
            this.nl = null;
            this.no = null;
            this.fd = 0;
            this.fw = 0;
            this.fm = 0;
            this.dc = 0;
            this.jk = 0;
            this.nr = null;
            this.jh = 0;
            this.jj = 0;
            this.jg = 0;
            this.nm = null;
            this.np = null;
            this.kk = null;
            this.d9 = 0;
            this.ea = 0;
            this.dj = 0;
            this.dk = 0;
            this.ki = null;
            this.kj = null;
            this.b3 = false;
            this.p = null;
            this.m = null;
            this.o = null;
            this.l = null;
            this.lx = null;
            this.jd = 0;
            this.je = 0;
            this.lz = null;
            this.ly = null;
            this.b4 = false;
            this.l0 = null;
            this.q = null;
            this.ex = 0;
            this.ey = 0;
            this.ez = 0;
            this.e0 = 0;
            this.b5 = false;
            this.l5 = null;
            this.b6 = false;
            this.e2 = 0;
            this.e3 = 0;
            this.e4 = 0;
            this.e5 = 0;
            this.l1 = null;
            this.l2 = null;
            this.e1 = 0;
            this.l3 = null;
            this.l4 = null;
            this.kr = null;
            this.n = null;
            this.kp = null;
            this.kq = null;
            this.ks = null;
            this.kt = null;
            this.ku = null;
            this.d0 = 0;
            this.d1 = 0;
            this.d2 = 0;
            this.d3 = 0;
            this.dv = 0;
            this.dw = 0;
            this.dx = 0;
            this.dy = 0;
            this.b2 = false;
            this.kv = null;
            this.dz = 0;
            this.kw = null;
            this.kx = null;
            this.kl = null;
            this.dr = 0;
            this.ds = 0;
            this.dt = 0;
            this.du = 0;
            this.dl = 0;
            this.dm = 0;
            this.dn = 0;
            this.dp = 0;
            this.b1 = false;
            this.km = null;
            this.dq = 0;
            this.kn = null;
            this.ko = null;
            this.k9 = null;
            this.la = null;
            this.eg = 0;
            this.lb = null;
            this.lc = null;
            this.ld = null;
            this.k8 = null;
            this.eh = 0;
            this.ei = 0;
            this.ej = 0;
            this.ek = 0;
            this.ec = 0;
            this.ed = 0;
            this.ee = 0;
            this.ef = 0;
            this.lj = null;
            this.lk = null;
            this.em = 0;
            this.ll = null;
            this.lm = null;
            this.k3 = null;
            this.k4 = null;
            this.k5 = null;
            this.eb = 0;
            this.k6 = null;
            this.k7 = null;
            this.le = null;
            this.lf = null;
            this.lg = null;
            this.el = 0;
            this.lh = null;
            this.li = null;
            this.df = 0;
            this.dg = 0;
            this.dh = 0;
            this.di = 0;
            this.kh = null;
            this.lr = null;
            this.ls = null;
            this.lt = null;
            this.et = 0;
            this.eu = 0;
            this.ev = 0;
            this.ew = 0;
            this.lu = null;
            this.es = 0;
            this.lv = null;
            this.lw = null;
            this.eo = 0;
            this.ep = 0;
            this.eq = 0;
            this.er = 0;
            this.ln = null;
            this.lo = null;
            this.en = 0;
            this.lp = null;
            this.lq = null;
            this.ky = null;
            this.kz = null;
            this.d5 = 0;
            this.d6 = 0;
            this.d7 = 0;
            this.d8 = 0;
            this.k0 = null;
            this.d4 = 0;
            this.k1 = null;
            this.k2 = null;
            this.aj = null;
            this.v = null;
            this.u = null;
            this.s = null;
            this.w = null;
            this.ak = null;
            this.m8 = null;
            this.mb = null;
            this.m7 = null;
            this.l9 = null;
            this.m6 = null;
            this.ai = null;
            this.t = null;
            this.ac = null;
            this.aa = null;
            this.z = null;
            this.y = null;
            this.ab = null;
            this.ad = null;
            this.mp = null;
            this.mc = null;
            this.mn = null;
            this.mm = null;
            this.mo = null;
            this.mq = null;
            this.kg = null;
            this.bl = null;
            this.md = null;
            this.bm = null;
            this.x = null;
            this.r = null;
            this.j = null;
            this.ah = null;
            this.h = null;
            this.i = null;
            this.mv = null;
            this.b9 = false;
            this.cb = false;
            this.ce = false;
            this.cg = false;
            this.mf = null;
            this.ca = false;
            this.me = null;
            this.mu = null;
            this.bp = null;
            this.j7 = null;
            this.nj = null;
            this.nn = null;
            this.nk = null;
            this.nt = null;
            this.fl = 0;
            this.fr = 0;
            this.mz = null;
            this.my = null;
            this.mx = null;
            this.fe = 0;
            this.ff = 0;
            this.ag = null;
            this.mw = null;
            this.ae = null;
            this.af = null;
            this.ji = 0;
            this.db = 0;
            this.bs = false;
            this.al = null;
            this.ns = null;
            this.fx = 0;
            this.br = false;
            this.fi = 0;
            this.fk = 0;
            this.fj = 0;
            this.fh = 0;
            this.j8 = null;
            this.nb = null;
            this.nc = null;
            this.mh = null;
            this.mg = null;
            this.mr = null;
            this.mt = null;
            this.ms = null;
            this.jl = 0;
            this.nq = null;
            this.kd = null;
            this.b0 = false;
            this.kf = null;
            this.de = 0;
            this.ke = null;
            this.bz = false;
            this.j9 = null;
            this.kb = null;
            this.ka = null;
            this.kc = null;
            this.jb = 0;
            this.jc = 0;
            this.cd = false;
            this.cf = false;
            this.b7 = false;
            this.l7 = null;
            this.l8 = null;
            this.jf = 0;
            this.bt = false;
            this.by = false;
            this.bu = false;
            this.ch = false;
            this.jy = null;
            this.bw = false;
            this.bx = false;
            this.bv = false;
            this.dd = 0;
            this.j0 = null;
            this.ja = 0;
            this.j2 = null;
            this.k = null;
            this.j5 = null;
            this.j6 = null;
            this.j3 = null;
            this.j1 = null;
            this.am = null;
            this.an = null;
            this.fy = 0;
            this.cc = false;
            this.mj = null;
            this.nv = null;
            this.ml = null;
            this.nx = null;
            this.mi = null;
            this.mk = null;
            this.fa = 0;
            this.e9 = 0;
            this.e6 = 0;
            this.e7 = 0;
            this.e8 = 0;
            this.fc = 0;
            this.fb = 0;
            this.b8 = false;
            this.nu = null;
            this.nw = null;
            this.f2 = 0;
            this.f1 = 0;
            this.fz = 0;
            this.f5 = 0;
            this.f0 = 0;
            this.f4 = 0;
            this.f3 = 0;
            this.ci = false;
            this.f7 = 0;
            this.f6 = 0;
            this.f9 = 0;
            this.f8 = 0;
            this.m9 = null;
            this.ni = null;
            this.ne = null;
            this.nf = null;
            this.ng = null;
            this.nd = null;
            this.nh = null;
            this.na = null;
            this.m1 = null;
            this.m2 = null;
            this.m3 = null;
            this.m0 = null;
            this.m4 = null;
            this.j4 = null;
            this.jz = null;
            this.m5 = null;
            this.ma = null;
            this.l6 = null;
        }
        get pixelScalingRatio() {
            return this.fg;
        }
        set pixelScalingRatio(a) {
            this.fg = a;
            this.g("PixelScalingRatio");
        }
        get titleLeftMargin() {
            return this.ft;
        }
        set titleLeftMargin(a) {
            this.ft = a;
            this.g("TitleLeftMargin");
        }
        get titleRightMargin() {
            return this.fu;
        }
        set titleRightMargin(a) {
            this.fu = a;
            this.g("TitleRightMargin");
        }
        get titleTopMargin() {
            return this.fv;
        }
        set titleTopMargin(a) {
            this.fv = a;
            this.g("TitleTopMargin");
        }
        get titleBottomMargin() {
            return this.fs;
        }
        set titleBottomMargin(a) {
            this.fs = a;
            this.g("TitleBottomMargin");
        }
        get subtitleLeftMargin() {
            return this.fo;
        }
        set subtitleLeftMargin(a) {
            this.fo = a;
            this.g("SubtitleLeftMargin");
        }
        get subtitleTopMargin() {
            return this.fq;
        }
        set subtitleTopMargin(a) {
            this.fq = a;
            this.g("SubtitleTopMargin");
        }
        get subtitleRightMargin() {
            return this.fp;
        }
        set subtitleRightMargin(a) {
            this.fp = a;
            this.g("SubtitleRightMargin");
        }
        get subtitleBottomMargin() {
            return this.fn;
        }
        set subtitleBottomMargin(a) {
            this.fn = a;
            this.g("SubtitleBottomMargin");
        }
        get subtitleTextColor() {
            return this.nl;
        }
        set subtitleTextColor(a) {
            this.nl = a;
            this.g("SubtitleTextColor");
        }
        get titleTextColor() {
            return this.no;
        }
        set titleTextColor(a) {
            this.no = a;
            this.g("TitleTextColor");
        }
        get leftMargin() {
            return this.fd;
        }
        set leftMargin(a) {
            this.fd = a;
            this.g("LeftMargin");
        }
        get topMargin() {
            return this.fw;
        }
        set topMargin(a) {
            this.fw = a;
            this.g("TopMargin");
        }
        get rightMargin() {
            return this.fm;
        }
        set rightMargin(a) {
            this.fm = a;
            this.g("RightMargin");
        }
        get bottomMargin() {
            return this.dc;
        }
        set bottomMargin(a) {
            this.dc = a;
            this.g("BottomMargin");
        }
        get transitionDuration() {
            return this.jk;
        }
        set transitionDuration(a) {
            this.jk = a;
            this.g("TransitionDuration");
        }
        get transitionEasingFunctionRef() {
            return this.nr;
        }
        set transitionEasingFunctionRef(a) {
            this.nr = a;
            this.g("TransitionEasingFunctionRef");
        }
        get highlightingTransitionDuration() {
            return this.jh;
        }
        set highlightingTransitionDuration(a) {
            this.jh = a;
            this.g("HighlightingTransitionDuration");
        }
        get selectionTransitionDuration() {
            return this.jj;
        }
        set selectionTransitionDuration(a) {
            this.jj = a;
            this.g("SelectionTransitionDuration");
        }
        get focusTransitionDuration() {
            return this.jg;
        }
        set focusTransitionDuration(a) {
            this.jg = a;
            this.g("FocusTransitionDuration");
        }
        get subtitleTextStyle() {
            return this.nm;
        }
        set subtitleTextStyle(a) {
            this.nm = a;
            this.g("SubtitleTextStyle");
        }
        get titleTextStyle() {
            return this.np;
        }
        set titleTextStyle(a) {
            this.np = a;
            this.g("TitleTextStyle");
        }
        get dataToolTipGroupingMode() {
            return this.kk;
        }
        set dataToolTipGroupingMode(a) {
            this.kk = a;
            this.g("DataToolTipGroupingMode");
        }
        get dataToolTipPositionOffsetX() {
            return this.d9;
        }
        set dataToolTipPositionOffsetX(a) {
            this.d9 = a;
            this.g("DataToolTipPositionOffsetX");
        }
        get dataToolTipPositionOffsetY() {
            return this.ea;
        }
        set dataToolTipPositionOffsetY(a) {
            this.ea = a;
            this.g("DataToolTipPositionOffsetY");
        }
        get dataToolTipDefaultPositionOffsetX() {
            return this.dj;
        }
        set dataToolTipDefaultPositionOffsetX(a) {
            this.dj = a;
            this.g("DataToolTipDefaultPositionOffsetX");
        }
        get dataToolTipDefaultPositionOffsetY() {
            return this.dk;
        }
        set dataToolTipDefaultPositionOffsetY(a) {
            this.dk = a;
            this.g("DataToolTipDefaultPositionOffsetY");
        }
        get dataToolTipGroupedPositionModeX() {
            return this.ki;
        }
        set dataToolTipGroupedPositionModeX(a) {
            this.ki = a;
            this.g("DataToolTipGroupedPositionModeX");
        }
        get dataToolTipGroupedPositionModeY() {
            return this.kj;
        }
        set dataToolTipGroupedPositionModeY(a) {
            this.kj = a;
            this.g("DataToolTipGroupedPositionModeY");
        }
        get dataToolTipShouldUpdateWhenSeriesDataChanges() {
            return this.b3;
        }
        set dataToolTipShouldUpdateWhenSeriesDataChanges(a) {
            this.b3 = a;
            this.g("DataToolTipShouldUpdateWhenSeriesDataChanges");
        }
        get dataToolTipIncludedSeries() {
            return this.p;
        }
        set dataToolTipIncludedSeries(a) {
            this.p = a;
            this.g("DataToolTipIncludedSeries");
        }
        get dataToolTipExcludedSeries() {
            return this.m;
        }
        set dataToolTipExcludedSeries(a) {
            this.m = a;
            this.g("DataToolTipExcludedSeries");
        }
        get dataToolTipIncludedColumns() {
            return this.o;
        }
        set dataToolTipIncludedColumns(a) {
            this.o = a;
            this.g("DataToolTipIncludedColumns");
        }
        get dataToolTipExcludedColumns() {
            return this.l;
        }
        set dataToolTipExcludedColumns(a) {
            this.l = a;
            this.g("DataToolTipExcludedColumns");
        }
        get dataToolTipValueFormatAbbreviation() {
            return this.lx;
        }
        set dataToolTipValueFormatAbbreviation(a) {
            this.lx = a;
            this.g("DataToolTipValueFormatAbbreviation");
        }
        get dataToolTipValueFormatMaxFractions() {
            return this.jd;
        }
        set dataToolTipValueFormatMaxFractions(a) {
            this.jd = a;
            this.g("DataToolTipValueFormatMaxFractions");
        }
        get dataToolTipValueFormatMinFractions() {
            return this.je;
        }
        set dataToolTipValueFormatMinFractions(a) {
            this.je = a;
            this.g("DataToolTipValueFormatMinFractions");
        }
        get dataToolTipValueFormatMode() {
            return this.lz;
        }
        set dataToolTipValueFormatMode(a) {
            this.lz = a;
            this.g("DataToolTipValueFormatMode");
        }
        get dataToolTipValueFormatCulture() {
            return this.ly;
        }
        set dataToolTipValueFormatCulture(a) {
            this.ly = a;
            this.g("DataToolTipValueFormatCulture");
        }
        get dataToolTipValueFormatUseGrouping() {
            return this.b4;
        }
        set dataToolTipValueFormatUseGrouping(a) {
            this.b4 = a;
            this.g("DataToolTipValueFormatUseGrouping");
        }
        get dataToolTipValueFormatString() {
            return this.l0;
        }
        set dataToolTipValueFormatString(a) {
            this.l0 = a;
            this.g("DataToolTipValueFormatString");
        }
        get dataToolTipValueFormatSpecifiers() {
            return this.q;
        }
        set dataToolTipValueFormatSpecifiers(a) {
            this.q = a;
            this.g("DataToolTipValueFormatSpecifiers");
        }
        get dataToolTipValueRowMarginBottom() {
            return this.ex;
        }
        set dataToolTipValueRowMarginBottom(a) {
            this.ex = a;
            this.g("DataToolTipValueRowMarginBottom");
        }
        get dataToolTipValueRowMarginLeft() {
            return this.ey;
        }
        set dataToolTipValueRowMarginLeft(a) {
            this.ey = a;
            this.g("DataToolTipValueRowMarginLeft");
        }
        get dataToolTipValueRowMarginRight() {
            return this.ez;
        }
        set dataToolTipValueRowMarginRight(a) {
            this.ez = a;
            this.g("DataToolTipValueRowMarginRight");
        }
        get dataToolTipValueRowMarginTop() {
            return this.e0;
        }
        set dataToolTipValueRowMarginTop(a) {
            this.e0 = a;
            this.g("DataToolTipValueRowMarginTop");
        }
        get dataToolTipValueRowVisible() {
            return this.b5;
        }
        set dataToolTipValueRowVisible(a) {
            this.b5 = a;
            this.g("DataToolTipValueRowVisible");
        }
        get dataToolTipValueTextWhenMissingData() {
            return this.l5;
        }
        set dataToolTipValueTextWhenMissingData(a) {
            this.l5 = a;
            this.g("DataToolTipValueTextWhenMissingData");
        }
        get dataToolTipValueTextUseSeriesColors() {
            return this.b6;
        }
        set dataToolTipValueTextUseSeriesColors(a) {
            this.b6 = a;
            this.g("DataToolTipValueTextUseSeriesColors");
        }
        get dataToolTipValueTextMarginBottom() {
            return this.e2;
        }
        set dataToolTipValueTextMarginBottom(a) {
            this.e2 = a;
            this.g("DataToolTipValueTextMarginBottom");
        }
        get dataToolTipValueTextMarginLeft() {
            return this.e3;
        }
        set dataToolTipValueTextMarginLeft(a) {
            this.e3 = a;
            this.g("DataToolTipValueTextMarginLeft");
        }
        get dataToolTipValueTextMarginRight() {
            return this.e4;
        }
        set dataToolTipValueTextMarginRight(a) {
            this.e4 = a;
            this.g("DataToolTipValueTextMarginRight");
        }
        get dataToolTipValueTextMarginTop() {
            return this.e5;
        }
        set dataToolTipValueTextMarginTop(a) {
            this.e5 = a;
            this.g("DataToolTipValueTextMarginTop");
        }
        get dataToolTipValueTextColor() {
            return this.l1;
        }
        set dataToolTipValueTextColor(a) {
            this.l1 = a;
            this.g("DataToolTipValueTextColor");
        }
        get dataToolTipValueTextFontFamily() {
            return this.l2;
        }
        set dataToolTipValueTextFontFamily(a) {
            this.l2 = a;
            this.g("DataToolTipValueTextFontFamily");
        }
        get dataToolTipValueTextFontSize() {
            return this.e1;
        }
        set dataToolTipValueTextFontSize(a) {
            this.e1 = a;
            this.g("DataToolTipValueTextFontSize");
        }
        get dataToolTipValueTextFontStyle() {
            return this.l3;
        }
        set dataToolTipValueTextFontStyle(a) {
            this.l3 = a;
            this.g("DataToolTipValueTextFontStyle");
        }
        get dataToolTipValueTextFontWeight() {
            return this.l4;
        }
        set dataToolTipValueTextFontWeight(a) {
            this.l4 = a;
            this.g("DataToolTipValueTextFontWeight");
        }
        get dataToolTipHeaderFormatString() {
            return this.kr;
        }
        set dataToolTipHeaderFormatString(a) {
            this.kr = a;
            this.g("DataToolTipHeaderFormatString");
        }
        get dataToolTipHeaderFormatSpecifiers() {
            return this.n;
        }
        set dataToolTipHeaderFormatSpecifiers(a) {
            this.n = a;
            this.g("DataToolTipHeaderFormatSpecifiers");
        }
        get dataToolTipHeaderFormatCulture() {
            return this.kp;
        }
        set dataToolTipHeaderFormatCulture(a) {
            this.kp = a;
            this.g("DataToolTipHeaderFormatCulture");
        }
        get dataToolTipHeaderFormatDate() {
            return this.kq;
        }
        set dataToolTipHeaderFormatDate(a) {
            this.kq = a;
            this.g("DataToolTipHeaderFormatDate");
        }
        get dataToolTipHeaderFormatTime() {
            return this.ks;
        }
        set dataToolTipHeaderFormatTime(a) {
            this.ks = a;
            this.g("DataToolTipHeaderFormatTime");
        }
        get dataToolTipHeaderText() {
            return this.kt;
        }
        set dataToolTipHeaderText(a) {
            this.kt = a;
            this.g("DataToolTipHeaderText");
        }
        get dataToolTipHeaderTextColor() {
            return this.ku;
        }
        set dataToolTipHeaderTextColor(a) {
            this.ku = a;
            this.g("DataToolTipHeaderTextColor");
        }
        get dataToolTipHeaderTextMarginBottom() {
            return this.d0;
        }
        set dataToolTipHeaderTextMarginBottom(a) {
            this.d0 = a;
            this.g("DataToolTipHeaderTextMarginBottom");
        }
        get dataToolTipHeaderTextMarginLeft() {
            return this.d1;
        }
        set dataToolTipHeaderTextMarginLeft(a) {
            this.d1 = a;
            this.g("DataToolTipHeaderTextMarginLeft");
        }
        get dataToolTipHeaderTextMarginRight() {
            return this.d2;
        }
        set dataToolTipHeaderTextMarginRight(a) {
            this.d2 = a;
            this.g("DataToolTipHeaderTextMarginRight");
        }
        get dataToolTipHeaderTextMarginTop() {
            return this.d3;
        }
        set dataToolTipHeaderTextMarginTop(a) {
            this.d3 = a;
            this.g("DataToolTipHeaderTextMarginTop");
        }
        get dataToolTipHeaderRowMarginBottom() {
            return this.dv;
        }
        set dataToolTipHeaderRowMarginBottom(a) {
            this.dv = a;
            this.g("DataToolTipHeaderRowMarginBottom");
        }
        get dataToolTipHeaderRowMarginLeft() {
            return this.dw;
        }
        set dataToolTipHeaderRowMarginLeft(a) {
            this.dw = a;
            this.g("DataToolTipHeaderRowMarginLeft");
        }
        get dataToolTipHeaderRowMarginRight() {
            return this.dx;
        }
        set dataToolTipHeaderRowMarginRight(a) {
            this.dx = a;
            this.g("DataToolTipHeaderRowMarginRight");
        }
        get dataToolTipHeaderRowMarginTop() {
            return this.dy;
        }
        set dataToolTipHeaderRowMarginTop(a) {
            this.dy = a;
            this.g("DataToolTipHeaderRowMarginTop");
        }
        get dataToolTipHeaderRowVisible() {
            return this.b2;
        }
        set dataToolTipHeaderRowVisible(a) {
            this.b2 = a;
            this.g("DataToolTipHeaderRowVisible");
        }
        get dataToolTipHeaderTextFontFamily() {
            return this.kv;
        }
        set dataToolTipHeaderTextFontFamily(a) {
            this.kv = a;
            this.g("DataToolTipHeaderTextFontFamily");
        }
        get dataToolTipHeaderTextFontSize() {
            return this.dz;
        }
        set dataToolTipHeaderTextFontSize(a) {
            this.dz = a;
            this.g("DataToolTipHeaderTextFontSize");
        }
        get dataToolTipHeaderTextFontStyle() {
            return this.kw;
        }
        set dataToolTipHeaderTextFontStyle(a) {
            this.kw = a;
            this.g("DataToolTipHeaderTextFontStyle");
        }
        get dataToolTipHeaderTextFontWeight() {
            return this.kx;
        }
        set dataToolTipHeaderTextFontWeight(a) {
            this.kx = a;
            this.g("DataToolTipHeaderTextFontWeight");
        }
        get dataToolTipGroupTextColor() {
            return this.kl;
        }
        set dataToolTipGroupTextColor(a) {
            this.kl = a;
            this.g("DataToolTipGroupTextColor");
        }
        get dataToolTipGroupTextMarginBottom() {
            return this.dr;
        }
        set dataToolTipGroupTextMarginBottom(a) {
            this.dr = a;
            this.g("DataToolTipGroupTextMarginBottom");
        }
        get dataToolTipGroupTextMarginLeft() {
            return this.ds;
        }
        set dataToolTipGroupTextMarginLeft(a) {
            this.ds = a;
            this.g("DataToolTipGroupTextMarginLeft");
        }
        get dataToolTipGroupTextMarginRight() {
            return this.dt;
        }
        set dataToolTipGroupTextMarginRight(a) {
            this.dt = a;
            this.g("DataToolTipGroupTextMarginRight");
        }
        get dataToolTipGroupTextMarginTop() {
            return this.du;
        }
        set dataToolTipGroupTextMarginTop(a) {
            this.du = a;
            this.g("DataToolTipGroupTextMarginTop");
        }
        get dataToolTipGroupRowMarginBottom() {
            return this.dl;
        }
        set dataToolTipGroupRowMarginBottom(a) {
            this.dl = a;
            this.g("DataToolTipGroupRowMarginBottom");
        }
        get dataToolTipGroupRowMarginLeft() {
            return this.dm;
        }
        set dataToolTipGroupRowMarginLeft(a) {
            this.dm = a;
            this.g("DataToolTipGroupRowMarginLeft");
        }
        get dataToolTipGroupRowMarginRight() {
            return this.dn;
        }
        set dataToolTipGroupRowMarginRight(a) {
            this.dn = a;
            this.g("DataToolTipGroupRowMarginRight");
        }
        get dataToolTipGroupRowMarginTop() {
            return this.dp;
        }
        set dataToolTipGroupRowMarginTop(a) {
            this.dp = a;
            this.g("DataToolTipGroupRowMarginTop");
        }
        get dataToolTipGroupRowVisible() {
            return this.b1;
        }
        set dataToolTipGroupRowVisible(a) {
            this.b1 = a;
            this.g("DataToolTipGroupRowVisible");
        }
        get dataToolTipGroupTextFontFamily() {
            return this.km;
        }
        set dataToolTipGroupTextFontFamily(a) {
            this.km = a;
            this.g("DataToolTipGroupTextFontFamily");
        }
        get dataToolTipGroupTextFontSize() {
            return this.dq;
        }
        set dataToolTipGroupTextFontSize(a) {
            this.dq = a;
            this.g("DataToolTipGroupTextFontSize");
        }
        get dataToolTipGroupTextFontStyle() {
            return this.kn;
        }
        set dataToolTipGroupTextFontStyle(a) {
            this.kn = a;
            this.g("DataToolTipGroupTextFontStyle");
        }
        get dataToolTipGroupTextFontWeight() {
            return this.ko;
        }
        set dataToolTipGroupTextFontWeight(a) {
            this.ko = a;
            this.g("DataToolTipGroupTextFontWeight");
        }
        get dataToolTipSummaryTitleTextColor() {
            return this.k9;
        }
        set dataToolTipSummaryTitleTextColor(a) {
            this.k9 = a;
            this.g("DataToolTipSummaryTitleTextColor");
        }
        get dataToolTipSummaryTitleTextFontFamily() {
            return this.la;
        }
        set dataToolTipSummaryTitleTextFontFamily(a) {
            this.la = a;
            this.g("DataToolTipSummaryTitleTextFontFamily");
        }
        get dataToolTipSummaryTitleTextFontSize() {
            return this.eg;
        }
        set dataToolTipSummaryTitleTextFontSize(a) {
            this.eg = a;
            this.g("DataToolTipSummaryTitleTextFontSize");
        }
        get dataToolTipSummaryTitleTextFontStyle() {
            return this.lb;
        }
        set dataToolTipSummaryTitleTextFontStyle(a) {
            this.lb = a;
            this.g("DataToolTipSummaryTitleTextFontStyle");
        }
        get dataToolTipSummaryTitleTextFontWeight() {
            return this.lc;
        }
        set dataToolTipSummaryTitleTextFontWeight(a) {
            this.lc = a;
            this.g("DataToolTipSummaryTitleTextFontWeight");
        }
        get dataToolTipSummaryType() {
            return this.ld;
        }
        set dataToolTipSummaryType(a) {
            this.ld = a;
            this.g("DataToolTipSummaryType");
        }
        get dataToolTipSummaryTitleText() {
            return this.k8;
        }
        set dataToolTipSummaryTitleText(a) {
            this.k8 = a;
            this.g("DataToolTipSummaryTitleText");
        }
        get dataToolTipSummaryTitleTextMarginBottom() {
            return this.eh;
        }
        set dataToolTipSummaryTitleTextMarginBottom(a) {
            this.eh = a;
            this.g("DataToolTipSummaryTitleTextMarginBottom");
        }
        get dataToolTipSummaryTitleTextMarginLeft() {
            return this.ei;
        }
        set dataToolTipSummaryTitleTextMarginLeft(a) {
            this.ei = a;
            this.g("DataToolTipSummaryTitleTextMarginLeft");
        }
        get dataToolTipSummaryTitleTextMarginRight() {
            return this.ej;
        }
        set dataToolTipSummaryTitleTextMarginRight(a) {
            this.ej = a;
            this.g("DataToolTipSummaryTitleTextMarginRight");
        }
        get dataToolTipSummaryTitleTextMarginTop() {
            return this.ek;
        }
        set dataToolTipSummaryTitleTextMarginTop(a) {
            this.ek = a;
            this.g("DataToolTipSummaryTitleTextMarginTop");
        }
        get dataToolTipSummaryRowMarginBottom() {
            return this.ec;
        }
        set dataToolTipSummaryRowMarginBottom(a) {
            this.ec = a;
            this.g("DataToolTipSummaryRowMarginBottom");
        }
        get dataToolTipSummaryRowMarginLeft() {
            return this.ed;
        }
        set dataToolTipSummaryRowMarginLeft(a) {
            this.ed = a;
            this.g("DataToolTipSummaryRowMarginLeft");
        }
        get dataToolTipSummaryRowMarginRight() {
            return this.ee;
        }
        set dataToolTipSummaryRowMarginRight(a) {
            this.ee = a;
            this.g("DataToolTipSummaryRowMarginRight");
        }
        get dataToolTipSummaryRowMarginTop() {
            return this.ef;
        }
        set dataToolTipSummaryRowMarginTop(a) {
            this.ef = a;
            this.g("DataToolTipSummaryRowMarginTop");
        }
        get dataToolTipSummaryValueTextColor() {
            return this.lj;
        }
        set dataToolTipSummaryValueTextColor(a) {
            this.lj = a;
            this.g("DataToolTipSummaryValueTextColor");
        }
        get dataToolTipSummaryValueTextFontFamily() {
            return this.lk;
        }
        set dataToolTipSummaryValueTextFontFamily(a) {
            this.lk = a;
            this.g("DataToolTipSummaryValueTextFontFamily");
        }
        get dataToolTipSummaryValueTextFontSize() {
            return this.em;
        }
        set dataToolTipSummaryValueTextFontSize(a) {
            this.em = a;
            this.g("DataToolTipSummaryValueTextFontSize");
        }
        get dataToolTipSummaryValueTextFontStyle() {
            return this.ll;
        }
        set dataToolTipSummaryValueTextFontStyle(a) {
            this.ll = a;
            this.g("DataToolTipSummaryValueTextFontStyle");
        }
        get dataToolTipSummaryValueTextFontWeight() {
            return this.lm;
        }
        set dataToolTipSummaryValueTextFontWeight(a) {
            this.lm = a;
            this.g("DataToolTipSummaryValueTextFontWeight");
        }
        get dataToolTipSummaryLabelText() {
            return this.k3;
        }
        set dataToolTipSummaryLabelText(a) {
            this.k3 = a;
            this.g("DataToolTipSummaryLabelText");
        }
        get dataToolTipSummaryLabelTextColor() {
            return this.k4;
        }
        set dataToolTipSummaryLabelTextColor(a) {
            this.k4 = a;
            this.g("DataToolTipSummaryLabelTextColor");
        }
        get dataToolTipSummaryLabelTextFontFamily() {
            return this.k5;
        }
        set dataToolTipSummaryLabelTextFontFamily(a) {
            this.k5 = a;
            this.g("DataToolTipSummaryLabelTextFontFamily");
        }
        get dataToolTipSummaryLabelTextFontSize() {
            return this.eb;
        }
        set dataToolTipSummaryLabelTextFontSize(a) {
            this.eb = a;
            this.g("DataToolTipSummaryLabelTextFontSize");
        }
        get dataToolTipSummaryLabelTextFontStyle() {
            return this.k6;
        }
        set dataToolTipSummaryLabelTextFontStyle(a) {
            this.k6 = a;
            this.g("DataToolTipSummaryLabelTextFontStyle");
        }
        get dataToolTipSummaryLabelTextFontWeight() {
            return this.k7;
        }
        set dataToolTipSummaryLabelTextFontWeight(a) {
            this.k7 = a;
            this.g("DataToolTipSummaryLabelTextFontWeight");
        }
        get dataToolTipSummaryUnitsText() {
            return this.le;
        }
        set dataToolTipSummaryUnitsText(a) {
            this.le = a;
            this.g("DataToolTipSummaryUnitsText");
        }
        get dataToolTipSummaryUnitsTextColor() {
            return this.lf;
        }
        set dataToolTipSummaryUnitsTextColor(a) {
            this.lf = a;
            this.g("DataToolTipSummaryUnitsTextColor");
        }
        get dataToolTipSummaryUnitsTextFontFamily() {
            return this.lg;
        }
        set dataToolTipSummaryUnitsTextFontFamily(a) {
            this.lg = a;
            this.g("DataToolTipSummaryUnitsTextFontFamily");
        }
        get dataToolTipSummaryUnitsTextFontSize() {
            return this.el;
        }
        set dataToolTipSummaryUnitsTextFontSize(a) {
            this.el = a;
            this.g("DataToolTipSummaryUnitsTextFontSize");
        }
        get dataToolTipSummaryUnitsTextFontStyle() {
            return this.lh;
        }
        set dataToolTipSummaryUnitsTextFontStyle(a) {
            this.lh = a;
            this.g("DataToolTipSummaryUnitsTextFontStyle");
        }
        get dataToolTipSummaryUnitsTextFontWeight() {
            return this.li;
        }
        set dataToolTipSummaryUnitsTextFontWeight(a) {
            this.li = a;
            this.g("DataToolTipSummaryUnitsTextFontWeight");
        }
        get dataToolTipBadgeMarginBottom() {
            return this.df;
        }
        set dataToolTipBadgeMarginBottom(a) {
            this.df = a;
            this.g("DataToolTipBadgeMarginBottom");
        }
        get dataToolTipBadgeMarginLeft() {
            return this.dg;
        }
        set dataToolTipBadgeMarginLeft(a) {
            this.dg = a;
            this.g("DataToolTipBadgeMarginLeft");
        }
        get dataToolTipBadgeMarginRight() {
            return this.dh;
        }
        set dataToolTipBadgeMarginRight(a) {
            this.dh = a;
            this.g("DataToolTipBadgeMarginRight");
        }
        get dataToolTipBadgeMarginTop() {
            return this.di;
        }
        set dataToolTipBadgeMarginTop(a) {
            this.di = a;
            this.g("DataToolTipBadgeMarginTop");
        }
        get dataToolTipBadgeShape() {
            return this.kh;
        }
        set dataToolTipBadgeShape(a) {
            this.kh = a;
            this.g("DataToolTipBadgeShape");
        }
        get dataToolTipUnitsDisplayMode() {
            return this.lr;
        }
        set dataToolTipUnitsDisplayMode(a) {
            this.lr = a;
            this.g("DataToolTipUnitsDisplayMode");
        }
        get dataToolTipUnitsText() {
            return this.ls;
        }
        set dataToolTipUnitsText(a) {
            this.ls = a;
            this.g("DataToolTipUnitsText");
        }
        get dataToolTipUnitsTextColor() {
            return this.lt;
        }
        set dataToolTipUnitsTextColor(a) {
            this.lt = a;
            this.g("DataToolTipUnitsTextColor");
        }
        get dataToolTipUnitsTextMarginBottom() {
            return this.et;
        }
        set dataToolTipUnitsTextMarginBottom(a) {
            this.et = a;
            this.g("DataToolTipUnitsTextMarginBottom");
        }
        get dataToolTipUnitsTextMarginLeft() {
            return this.eu;
        }
        set dataToolTipUnitsTextMarginLeft(a) {
            this.eu = a;
            this.g("DataToolTipUnitsTextMarginLeft");
        }
        get dataToolTipUnitsTextMarginRight() {
            return this.ev;
        }
        set dataToolTipUnitsTextMarginRight(a) {
            this.ev = a;
            this.g("DataToolTipUnitsTextMarginRight");
        }
        get dataToolTipUnitsTextMarginTop() {
            return this.ew;
        }
        set dataToolTipUnitsTextMarginTop(a) {
            this.ew = a;
            this.g("DataToolTipUnitsTextMarginTop");
        }
        get dataToolTipUnitsTextFontFamily() {
            return this.lu;
        }
        set dataToolTipUnitsTextFontFamily(a) {
            this.lu = a;
            this.g("DataToolTipUnitsTextFontFamily");
        }
        get dataToolTipUnitsTextFontSize() {
            return this.es;
        }
        set dataToolTipUnitsTextFontSize(a) {
            this.es = a;
            this.g("DataToolTipUnitsTextFontSize");
        }
        get dataToolTipUnitsTextFontStyle() {
            return this.lv;
        }
        set dataToolTipUnitsTextFontStyle(a) {
            this.lv = a;
            this.g("DataToolTipUnitsTextFontStyle");
        }
        get dataToolTipUnitsTextFontWeight() {
            return this.lw;
        }
        set dataToolTipUnitsTextFontWeight(a) {
            this.lw = a;
            this.g("DataToolTipUnitsTextFontWeight");
        }
        get dataToolTipTitleTextMarginBottom() {
            return this.eo;
        }
        set dataToolTipTitleTextMarginBottom(a) {
            this.eo = a;
            this.g("DataToolTipTitleTextMarginBottom");
        }
        get dataToolTipTitleTextMarginLeft() {
            return this.ep;
        }
        set dataToolTipTitleTextMarginLeft(a) {
            this.ep = a;
            this.g("DataToolTipTitleTextMarginLeft");
        }
        get dataToolTipTitleTextMarginRight() {
            return this.eq;
        }
        set dataToolTipTitleTextMarginRight(a) {
            this.eq = a;
            this.g("DataToolTipTitleTextMarginRight");
        }
        get dataToolTipTitleTextMarginTop() {
            return this.er;
        }
        set dataToolTipTitleTextMarginTop(a) {
            this.er = a;
            this.g("DataToolTipTitleTextMarginTop");
        }
        get dataToolTipTitleTextColor() {
            return this.ln;
        }
        set dataToolTipTitleTextColor(a) {
            this.ln = a;
            this.g("DataToolTipTitleTextColor");
        }
        get dataToolTipTitleTextFontFamily() {
            return this.lo;
        }
        set dataToolTipTitleTextFontFamily(a) {
            this.lo = a;
            this.g("DataToolTipTitleTextFontFamily");
        }
        get dataToolTipTitleTextFontSize() {
            return this.en;
        }
        set dataToolTipTitleTextFontSize(a) {
            this.en = a;
            this.g("DataToolTipTitleTextFontSize");
        }
        get dataToolTipTitleTextFontStyle() {
            return this.lp;
        }
        set dataToolTipTitleTextFontStyle(a) {
            this.lp = a;
            this.g("DataToolTipTitleTextFontStyle");
        }
        get dataToolTipTitleTextFontWeight() {
            return this.lq;
        }
        set dataToolTipTitleTextFontWeight(a) {
            this.lq = a;
            this.g("DataToolTipTitleTextFontWeight");
        }
        get dataToolTipLabelDisplayMode() {
            return this.ky;
        }
        set dataToolTipLabelDisplayMode(a) {
            this.ky = a;
            this.g("DataToolTipLabelDisplayMode");
        }
        get dataToolTipLabelTextColor() {
            return this.kz;
        }
        set dataToolTipLabelTextColor(a) {
            this.kz = a;
            this.g("DataToolTipLabelTextColor");
        }
        get dataToolTipLabelTextMarginBottom() {
            return this.d5;
        }
        set dataToolTipLabelTextMarginBottom(a) {
            this.d5 = a;
            this.g("DataToolTipLabelTextMarginBottom");
        }
        get dataToolTipLabelTextMarginLeft() {
            return this.d6;
        }
        set dataToolTipLabelTextMarginLeft(a) {
            this.d6 = a;
            this.g("DataToolTipLabelTextMarginLeft");
        }
        get dataToolTipLabelTextMarginRight() {
            return this.d7;
        }
        set dataToolTipLabelTextMarginRight(a) {
            this.d7 = a;
            this.g("DataToolTipLabelTextMarginRight");
        }
        get dataToolTipLabelTextMarginTop() {
            return this.d8;
        }
        set dataToolTipLabelTextMarginTop(a) {
            this.d8 = a;
            this.g("DataToolTipLabelTextMarginTop");
        }
        get dataToolTipLabelTextFontFamily() {
            return this.k0;
        }
        set dataToolTipLabelTextFontFamily(a) {
            this.k0 = a;
            this.g("DataToolTipLabelTextFontFamily");
        }
        get dataToolTipLabelTextFontSize() {
            return this.d4;
        }
        set dataToolTipLabelTextFontSize(a) {
            this.d4 = a;
            this.g("DataToolTipLabelTextFontSize");
        }
        get dataToolTipLabelTextFontStyle() {
            return this.k1;
        }
        set dataToolTipLabelTextFontStyle(a) {
            this.k1 = a;
            this.g("DataToolTipLabelTextFontStyle");
        }
        get dataToolTipLabelTextFontWeight() {
            return this.k2;
        }
        set dataToolTipLabelTextFontWeight(a) {
            this.k2 = a;
            this.g("DataToolTipLabelTextFontWeight");
        }
        get sortDescriptions() {
            return this.aj;
        }
        set sortDescriptions(a) {
            this.aj = a;
            this.g("SortDescriptions");
        }
        get groupSortDescriptions() {
            return this.v;
        }
        set groupSortDescriptions(a) {
            this.v = a;
            this.g("GroupSortDescriptions");
        }
        get groupDescriptions() {
            return this.u;
        }
        set groupDescriptions(a) {
            this.u = a;
            this.g("GroupDescriptions");
        }
        get filterExpressions() {
            return this.s;
        }
        set filterExpressions(a) {
            this.s = a;
            this.g("FilterExpressions");
        }
        get highlightFilterExpressions() {
            return this.w;
        }
        set highlightFilterExpressions(a) {
            this.w = a;
            this.g("HighlightFilterExpressions");
        }
        get summaryDescriptions() {
            return this.ak;
        }
        set summaryDescriptions(a) {
            this.ak = a;
            this.g("SummaryDescriptions");
        }
        get selectionMode() {
            return this.m8;
        }
        set selectionMode(a) {
            this.m8 = a;
            this.g("SelectionMode");
        }
        get focusMode() {
            return this.mb;
        }
        set focusMode(a) {
            this.mb = a;
            this.g("FocusMode");
        }
        get selectionBrush() {
            return this.m7;
        }
        set selectionBrush(a) {
            this.m7 = a;
            this.g("SelectionBrush");
        }
        get focusBrush() {
            return this.l9;
        }
        set focusBrush(a) {
            this.l9 = a;
            this.g("FocusBrush");
        }
        get selectionBehavior() {
            return this.m6;
        }
        set selectionBehavior(a) {
            this.m6 = a;
            this.g("SelectionBehavior");
        }
        get selectedSeriesItems() {
            return this.ai;
        }
        set selectedSeriesItems(a) {
            this.ai = a;
            this.g("SelectedSeriesItems");
        }
        get focusedSeriesItems() {
            return this.t;
        }
        set focusedSeriesItems(a) {
            this.t = a;
            this.g("FocusedSeriesItems");
        }
        get initialSortDescriptions() {
            return this.ac;
        }
        set initialSortDescriptions(a) {
            this.ac = a;
            this.g("InitialSortDescriptions");
        }
        get initialGroupSortDescriptions() {
            return this.aa;
        }
        set initialGroupSortDescriptions(a) {
            this.aa = a;
            this.g("InitialGroupSortDescriptions");
        }
        get initialGroupDescriptions() {
            return this.z;
        }
        set initialGroupDescriptions(a) {
            this.z = a;
            this.g("InitialGroupDescriptions");
        }
        get initialFilterExpressions() {
            return this.y;
        }
        set initialFilterExpressions(a) {
            this.y = a;
            this.g("InitialFilterExpressions");
        }
        get initialHighlightFilterExpressions() {
            return this.ab;
        }
        set initialHighlightFilterExpressions(a) {
            this.ab = a;
            this.g("InitialHighlightFilterExpressions");
        }
        get initialSummaryDescriptions() {
            return this.ad;
        }
        set initialSummaryDescriptions(a) {
            this.ad = a;
            this.g("InitialSummaryDescriptions");
        }
        get initialSorts() {
            return this.mp;
        }
        set initialSorts(a) {
            this.mp = a;
            this.g("InitialSorts");
        }
        get groupSorts() {
            return this.mc;
        }
        set groupSorts(a) {
            this.mc = a;
            this.g("GroupSorts");
        }
        get initialGroups() {
            return this.mn;
        }
        set initialGroups(a) {
            this.mn = a;
            this.g("InitialGroups");
        }
        get initialFilter() {
            return this.mm;
        }
        set initialFilter(a) {
            this.mm = a;
            this.g("InitialFilter");
        }
        get initialHighlightFilter() {
            return this.mo;
        }
        set initialHighlightFilter(a) {
            this.mo = a;
            this.g("InitialHighlightFilter");
        }
        get initialSummaries() {
            return this.mq;
        }
        set initialSummaries(a) {
            this.mq = a;
            this.g("InitialSummaries");
        }
        get dataSourceRef() {
            return this.kg;
        }
        set dataSourceRef(a) {
            this.kg = a;
            this.g("DataSourceRef");
        }
        get dataSource() {
            return this.bl;
        }
        set dataSource(a) {
            this.bl = a;
            this.g("DataSource");
        }
        get highlightedDataSourceRef() {
            return this.md;
        }
        set highlightedDataSourceRef(a) {
            this.md = a;
            this.g("HighlightedDataSourceRef");
        }
        get highlightedDataSource() {
            return this.bm;
        }
        set highlightedDataSource(a) {
            this.bm = a;
            this.g("HighlightedDataSource");
        }
        get includedProperties() {
            return this.x;
        }
        set includedProperties(a) {
            this.x = a;
            this.g("IncludedProperties");
        }
        get excludedProperties() {
            return this.r;
        }
        set excludedProperties(a) {
            this.r = a;
            this.g("ExcludedProperties");
        }
        get brushes() {
            return this.j;
        }
        set brushes(a) {
            this.j = a;
            this.g("Brushes");
        }
        get outlines() {
            return this.ah;
        }
        set outlines(a) {
            this.ah = a;
            this.g("Outlines");
        }
        get actualBrushes() {
            return this.h;
        }
        set actualBrushes(a) {
            this.h = a;
            this.g("ActualBrushes");
        }
        get actualOutlines() {
            return this.i;
        }
        set actualOutlines(a) {
            this.i = a;
            this.g("ActualOutlines");
        }
        get legendRef() {
            return this.mv;
        }
        set legendRef(a) {
            this.mv = a;
            this.g("LegendRef");
        }
        get isHorizontalZoomEnabled() {
            return this.b9;
        }
        set isHorizontalZoomEnabled(a) {
            this.b9 = a;
            this.g("IsHorizontalZoomEnabled");
        }
        get isVerticalZoomEnabled() {
            return this.cb;
        }
        set isVerticalZoomEnabled(a) {
            this.cb = a;
            this.g("IsVerticalZoomEnabled");
        }
        get shouldDisplayMockData() {
            return this.ce;
        }
        set shouldDisplayMockData(a) {
            this.ce = a;
            this.g("ShouldDisplayMockData");
        }
        get shouldUseSkeletonStyleForMockData() {
            return this.cg;
        }
        set shouldUseSkeletonStyleForMockData(a) {
            this.cg = a;
            this.g("ShouldUseSkeletonStyleForMockData");
        }
        get highlightedValuesDisplayMode() {
            return this.mf;
        }
        set highlightedValuesDisplayMode(a) {
            this.mf = a;
            this.g("HighlightedValuesDisplayMode");
        }
        get isSeriesHighlightingEnabled() {
            return this.ca;
        }
        set isSeriesHighlightingEnabled(a) {
            this.ca = a;
            this.g("IsSeriesHighlightingEnabled");
        }
        get highlightedLegendItemVisibility() {
            return this.me;
        }
        set highlightedLegendItemVisibility(a) {
            this.me = a;
            this.g("HighlightedLegendItemVisibility");
        }
        get legendItemVisibility() {
            return this.mu;
        }
        set legendItemVisibility(a) {
            this.mu = a;
            this.g("LegendItemVisibility");
        }
        get windowRect() {
            return this.bp;
        }
        set windowRect(a) {
            this.bp = a;
            this.g("WindowRect");
        }
        get chartTitle() {
            return this.j7;
        }
        set chartTitle(a) {
            this.j7 = a;
            this.g("ChartTitle");
        }
        get subtitle() {
            return this.nj;
        }
        set subtitle(a) {
            this.nj = a;
            this.g("Subtitle");
        }
        get titleAlignment() {
            return this.nn;
        }
        set titleAlignment(a) {
            this.nn = a;
            this.g("TitleAlignment");
        }
        get subtitleAlignment() {
            return this.nk;
        }
        set subtitleAlignment(a) {
            this.nk = a;
            this.g("SubtitleAlignment");
        }
        get unknownValuePlotting() {
            return this.nt;
        }
        set unknownValuePlotting(a) {
            this.nt = a;
            this.g("UnknownValuePlotting");
        }
        get resolution() {
            return this.fl;
        }
        set resolution(a) {
            this.fl = a;
            this.g("Resolution");
        }
        get thickness() {
            return this.fr;
        }
        set thickness(a) {
            this.fr = a;
            this.g("Thickness");
        }
        get outlineMode() {
            return this.mz;
        }
        set outlineMode(a) {
            this.mz = a;
            this.g("OutlineMode");
        }
        get markerOutlineMode() {
            return this.my;
        }
        set markerOutlineMode(a) {
            this.my = a;
            this.g("MarkerOutlineMode");
        }
        get markerFillMode() {
            return this.mx;
        }
        set markerFillMode(a) {
            this.mx = a;
            this.g("MarkerFillMode");
        }
        get markerFillOpacity() {
            return this.fe;
        }
        set markerFillOpacity(a) {
            this.fe = a;
            this.g("MarkerFillOpacity");
        }
        get markerThickness() {
            return this.ff;
        }
        set markerThickness(a) {
            this.ff = a;
            this.g("MarkerThickness");
        }
        get markerTypes() {
            return this.ag;
        }
        set markerTypes(a) {
            this.ag = a;
            this.g("MarkerTypes");
        }
        get markerAutomaticBehavior() {
            return this.mw;
        }
        set markerAutomaticBehavior(a) {
            this.mw = a;
            this.g("MarkerAutomaticBehavior");
        }
        get markerBrushes() {
            return this.ae;
        }
        set markerBrushes(a) {
            this.ae = a;
            this.g("MarkerBrushes");
        }
        get markerOutlines() {
            return this.af;
        }
        set markerOutlines(a) {
            this.af = a;
            this.g("MarkerOutlines");
        }
        get markerMaxCount() {
            return this.ji;
        }
        set markerMaxCount(a) {
            this.ji = a;
            this.g("MarkerMaxCount");
        }
        get areaFillOpacity() {
            return this.db;
        }
        set areaFillOpacity(a) {
            this.db = a;
            this.g("AreaFillOpacity");
        }
        get animateSeriesWhenAxisRangeChanges() {
            return this.bs;
        }
        set animateSeriesWhenAxisRangeChanges(a) {
            this.bs = a;
            this.g("AnimateSeriesWhenAxisRangeChanges");
        }
        get trendLineBrushes() {
            return this.al;
        }
        set trendLineBrushes(a) {
            this.al = a;
            this.g("TrendLineBrushes");
        }
        get trendLineType() {
            return this.ns;
        }
        set trendLineType(a) {
            this.ns = a;
            this.g("TrendLineType");
        }
        get trendLineThickness() {
            return this.fx;
        }
        set trendLineThickness(a) {
            this.fx = a;
            this.g("TrendLineThickness");
        }
        get alignsGridLinesToPixels() {
            return this.br;
        }
        set alignsGridLinesToPixels(a) {
            this.br = a;
            this.g("AlignsGridLinesToPixels");
        }
        get plotAreaMarginLeft() {
            return this.fi;
        }
        set plotAreaMarginLeft(a) {
            this.fi = a;
            this.g("PlotAreaMarginLeft");
        }
        get plotAreaMarginTop() {
            return this.fk;
        }
        set plotAreaMarginTop(a) {
            this.fk = a;
            this.g("PlotAreaMarginTop");
        }
        get plotAreaMarginRight() {
            return this.fj;
        }
        set plotAreaMarginRight(a) {
            this.fj = a;
            this.g("PlotAreaMarginRight");
        }
        get plotAreaMarginBottom() {
            return this.fh;
        }
        set plotAreaMarginBottom(a) {
            this.fh = a;
            this.g("PlotAreaMarginBottom");
        }
        get computedPlotAreaMarginMode() {
            return this.j8;
        }
        set computedPlotAreaMarginMode(a) {
            this.j8 = a;
            this.g("ComputedPlotAreaMarginMode");
        }
        get seriesPlotAreaMarginHorizontalMode() {
            return this.nb;
        }
        set seriesPlotAreaMarginHorizontalMode(a) {
            this.nb = a;
            this.g("SeriesPlotAreaMarginHorizontalMode");
        }
        get seriesPlotAreaMarginVerticalMode() {
            return this.nc;
        }
        set seriesPlotAreaMarginVerticalMode(a) {
            this.nc = a;
            this.g("SeriesPlotAreaMarginVerticalMode");
        }
        get highlightingMode() {
            return this.mh;
        }
        set highlightingMode(a) {
            this.mh = a;
            this.g("HighlightingMode");
        }
        get highlightingBehavior() {
            return this.mg;
        }
        set highlightingBehavior(a) {
            this.mg = a;
            this.g("HighlightingBehavior");
        }
        get legendHighlightingMode() {
            return this.mr;
        }
        set legendHighlightingMode(a) {
            this.mr = a;
            this.g("LegendHighlightingMode");
        }
        get legendItemBadgeShape() {
            return this.mt;
        }
        set legendItemBadgeShape(a) {
            this.mt = a;
            this.g("LegendItemBadgeShape");
        }
        get legendItemBadgeMode() {
            return this.ms;
        }
        set legendItemBadgeMode(a) {
            this.ms = a;
            this.g("LegendItemBadgeMode");
        }
        get trendLinePeriod() {
            return this.jl;
        }
        set trendLinePeriod(a) {
            this.jl = a;
            this.g("TrendLinePeriod");
        }
        get toolTipType() {
            return this.nq;
        }
        set toolTipType(a) {
            this.nq = a;
            this.g("ToolTipType");
        }
        get crosshairsDisplayMode() {
            return this.kd;
        }
        set crosshairsDisplayMode(a) {
            this.kd = a;
            this.g("CrosshairsDisplayMode");
        }
        get crosshairsSnapToData() {
            return this.b0;
        }
        set crosshairsSnapToData(a) {
            this.b0 = a;
            this.g("CrosshairsSnapToData");
        }
        get crosshairsLineVerticalStroke() {
            return this.kf;
        }
        set crosshairsLineVerticalStroke(a) {
            this.kf = a;
            this.g("CrosshairsLineVerticalStroke");
        }
        get crosshairsLineThickness() {
            return this.de;
        }
        set crosshairsLineThickness(a) {
            this.de = a;
            this.g("CrosshairsLineThickness");
        }
        get crosshairsLineHorizontalStroke() {
            return this.ke;
        }
        set crosshairsLineHorizontalStroke(a) {
            this.ke = a;
            this.g("CrosshairsLineHorizontalStroke");
        }
        get crosshairsAnnotationEnabled() {
            return this.bz;
        }
        set crosshairsAnnotationEnabled(a) {
            this.bz = a;
            this.g("CrosshairsAnnotationEnabled");
        }
        get crosshairsAnnotationXAxisBackground() {
            return this.j9;
        }
        set crosshairsAnnotationXAxisBackground(a) {
            this.j9 = a;
            this.g("CrosshairsAnnotationXAxisBackground");
        }
        get crosshairsAnnotationYAxisBackground() {
            return this.kb;
        }
        set crosshairsAnnotationYAxisBackground(a) {
            this.kb = a;
            this.g("CrosshairsAnnotationYAxisBackground");
        }
        get crosshairsAnnotationXAxisTextColor() {
            return this.ka;
        }
        set crosshairsAnnotationXAxisTextColor(a) {
            this.ka = a;
            this.g("CrosshairsAnnotationXAxisTextColor");
        }
        get crosshairsAnnotationYAxisTextColor() {
            return this.kc;
        }
        set crosshairsAnnotationYAxisTextColor(a) {
            this.kc = a;
            this.g("CrosshairsAnnotationYAxisTextColor");
        }
        get crosshairsAnnotationXAxisPrecision() {
            return this.jb;
        }
        set crosshairsAnnotationXAxisPrecision(a) {
            this.jb = a;
            this.g("CrosshairsAnnotationXAxisPrecision");
        }
        get crosshairsAnnotationYAxisPrecision() {
            return this.jc;
        }
        set crosshairsAnnotationYAxisPrecision(a) {
            this.jc = a;
            this.g("CrosshairsAnnotationYAxisPrecision");
        }
        get shouldAvoidAxisAnnotationCollisions() {
            return this.cd;
        }
        set shouldAvoidAxisAnnotationCollisions(a) {
            this.cd = a;
            this.g("ShouldAvoidAxisAnnotationCollisions");
        }
        get shouldPanOnMaximumZoom() {
            return this.cf;
        }
        set shouldPanOnMaximumZoom(a) {
            this.cf = a;
            this.g("ShouldPanOnMaximumZoom");
        }
        get finalValueAnnotationsVisible() {
            return this.b7;
        }
        set finalValueAnnotationsVisible(a) {
            this.b7 = a;
            this.g("FinalValueAnnotationsVisible");
        }
        get finalValueAnnotationsBackground() {
            return this.l7;
        }
        set finalValueAnnotationsBackground(a) {
            this.l7 = a;
            this.g("FinalValueAnnotationsBackground");
        }
        get finalValueAnnotationsTextColor() {
            return this.l8;
        }
        set finalValueAnnotationsTextColor(a) {
            this.l8 = a;
            this.g("FinalValueAnnotationsTextColor");
        }
        get finalValueAnnotationsPrecision() {
            return this.jf;
        }
        set finalValueAnnotationsPrecision(a) {
            this.jf = a;
            this.g("FinalValueAnnotationsPrecision");
        }
        get autoCalloutsVisible() {
            return this.bt;
        }
        set autoCalloutsVisible(a) {
            this.bt = a;
            this.g("AutoCalloutsVisible");
        }
        get calloutsVisible() {
            return this.by;
        }
        set calloutsVisible(a) {
            this.by = a;
            this.g("CalloutsVisible");
        }
        get calloutStyleUpdatingEventEnabled() {
            return this.bu;
        }
        set calloutStyleUpdatingEventEnabled(a) {
            this.bu = a;
            this.g("CalloutStyleUpdatingEventEnabled");
        }
        get useValueForAutoCalloutLabels() {
            return this.ch;
        }
        set useValueForAutoCalloutLabels(a) {
            this.ch = a;
            this.g("UseValueForAutoCalloutLabels");
        }
        get calloutCollisionMode() {
            return this.jy;
        }
        set calloutCollisionMode(a) {
            this.jy = a;
            this.g("CalloutCollisionMode");
        }
        get calloutsUseItemColorForFill() {
            return this.bw;
        }
        set calloutsUseItemColorForFill(a) {
            this.bw = a;
            this.g("CalloutsUseItemColorForFill");
        }
        get calloutsUseItemColorForOutline() {
            return this.bx;
        }
        set calloutsUseItemColorForOutline(a) {
            this.bx = a;
            this.g("CalloutsUseItemColorForOutline");
        }
        get calloutsUseAutoContrastingLabelColors() {
            return this.bv;
        }
        set calloutsUseAutoContrastingLabelColors(a) {
            this.bv = a;
            this.g("CalloutsUseAutoContrastingLabelColors");
        }
        get calloutsStrokeThickness() {
            return this.dd;
        }
        set calloutsStrokeThickness(a) {
            this.dd = a;
            this.g("CalloutsStrokeThickness");
        }
        get calloutsBackground() {
            return this.j0;
        }
        set calloutsBackground(a) {
            this.j0 = a;
            this.g("CalloutsBackground");
        }
        get calloutsAutoLabelPrecision() {
            return this.ja;
        }
        set calloutsAutoLabelPrecision(a) {
            this.ja = a;
            this.g("CalloutsAutoLabelPrecision");
        }
        get calloutsDataSourceRef() {
            return this.j2;
        }
        set calloutsDataSourceRef(a) {
            this.j2 = a;
            this.g("CalloutsDataSourceRef");
        }
        get calloutsAllowedPositions() {
            return this.k;
        }
        set calloutsAllowedPositions(a) {
            this.k = a;
            this.g("CalloutsAllowedPositions");
        }
        get calloutsXMemberPath() {
            return this.j5;
        }
        set calloutsXMemberPath(a) {
            this.j5 = a;
            this.g("CalloutsXMemberPath");
        }
        get calloutsYMemberPath() {
            return this.j6;
        }
        set calloutsYMemberPath(a) {
            this.j6 = a;
            this.g("CalloutsYMemberPath");
        }
        get calloutsLabelMemberPath() {
            return this.j3;
        }
        set calloutsLabelMemberPath(a) {
            this.j3 = a;
            this.g("CalloutsLabelMemberPath");
        }
        get calloutsContentMemberPath() {
            return this.j1;
        }
        set calloutsContentMemberPath(a) {
            this.j1 = a;
            this.g("CalloutsContentMemberPath");
        }
        get valueLines() {
            return this.am;
        }
        set valueLines(a) {
            this.am = a;
            this.g("ValueLines");
        }
        get valueLinesBrushes() {
            return this.an;
        }
        set valueLinesBrushes(a) {
            this.an = a;
            this.g("ValueLinesBrushes");
        }
        get valueLinesThickness() {
            return this.fy;
        }
        set valueLinesThickness(a) {
            this.fy = a;
            this.g("ValueLinesThickness");
        }
        get seriesValueLayerUseLegend() {
            return this.cc;
        }
        set seriesValueLayerUseLegend(a) {
            this.cc = a;
            this.g("SeriesValueLayerUseLegend");
        }
        get horizontalViewScrollbarMode() {
            return this.mj;
        }
        set horizontalViewScrollbarMode(a) {
            this.mj = a;
            this.g("HorizontalViewScrollbarMode");
        }
        get verticalViewScrollbarMode() {
            return this.nv;
        }
        set verticalViewScrollbarMode(a) {
            this.nv = a;
            this.g("VerticalViewScrollbarMode");
        }
        get horizontalViewScrollbarPosition() {
            return this.ml;
        }
        set horizontalViewScrollbarPosition(a) {
            this.ml = a;
            this.g("HorizontalViewScrollbarPosition");
        }
        get verticalViewScrollbarPosition() {
            return this.nx;
        }
        set verticalViewScrollbarPosition(a) {
            this.nx = a;
            this.g("VerticalViewScrollbarPosition");
        }
        get horizontalViewScrollbarFill() {
            return this.mi;
        }
        set horizontalViewScrollbarFill(a) {
            this.mi = a;
            this.g("HorizontalViewScrollbarFill");
        }
        get horizontalViewScrollbarOutline() {
            return this.mk;
        }
        set horizontalViewScrollbarOutline(a) {
            this.mk = a;
            this.g("HorizontalViewScrollbarOutline");
        }
        get horizontalViewScrollbarStrokeThickness() {
            return this.fa;
        }
        set horizontalViewScrollbarStrokeThickness(a) {
            this.fa = a;
            this.g("HorizontalViewScrollbarStrokeThickness");
        }
        get horizontalViewScrollbarMaxOpacity() {
            return this.e9;
        }
        set horizontalViewScrollbarMaxOpacity(a) {
            this.e9 = a;
            this.g("HorizontalViewScrollbarMaxOpacity");
        }
        get horizontalViewScrollbarCornerRadius() {
            return this.e6;
        }
        set horizontalViewScrollbarCornerRadius(a) {
            this.e6 = a;
            this.g("HorizontalViewScrollbarCornerRadius");
        }
        get horizontalViewScrollbarHeight() {
            return this.e7;
        }
        set horizontalViewScrollbarHeight(a) {
            this.e7 = a;
            this.g("HorizontalViewScrollbarHeight");
        }
        get horizontalViewScrollbarInset() {
            return this.e8;
        }
        set horizontalViewScrollbarInset(a) {
            this.e8 = a;
            this.g("HorizontalViewScrollbarInset");
        }
        get horizontalViewScrollbarTrackStartInset() {
            return this.fc;
        }
        set horizontalViewScrollbarTrackStartInset(a) {
            this.fc = a;
            this.g("HorizontalViewScrollbarTrackStartInset");
        }
        get horizontalViewScrollbarTrackEndInset() {
            return this.fb;
        }
        set horizontalViewScrollbarTrackEndInset(a) {
            this.fb = a;
            this.g("HorizontalViewScrollbarTrackEndInset");
        }
        get horizontalViewScrollbarShouldAddAutoTrackInsets() {
            return this.b8;
        }
        set horizontalViewScrollbarShouldAddAutoTrackInsets(a) {
            this.b8 = a;
            this.g("HorizontalViewScrollbarShouldAddAutoTrackInsets");
        }
        get verticalViewScrollbarFill() {
            return this.nu;
        }
        set verticalViewScrollbarFill(a) {
            this.nu = a;
            this.g("VerticalViewScrollbarFill");
        }
        get verticalViewScrollbarOutline() {
            return this.nw;
        }
        set verticalViewScrollbarOutline(a) {
            this.nw = a;
            this.g("VerticalViewScrollbarOutline");
        }
        get verticalViewScrollbarStrokeThickness() {
            return this.f2;
        }
        set verticalViewScrollbarStrokeThickness(a) {
            this.f2 = a;
            this.g("VerticalViewScrollbarStrokeThickness");
        }
        get verticalViewScrollbarMaxOpacity() {
            return this.f1;
        }
        set verticalViewScrollbarMaxOpacity(a) {
            this.f1 = a;
            this.g("VerticalViewScrollbarMaxOpacity");
        }
        get verticalViewScrollbarCornerRadius() {
            return this.fz;
        }
        set verticalViewScrollbarCornerRadius(a) {
            this.fz = a;
            this.g("VerticalViewScrollbarCornerRadius");
        }
        get verticalViewScrollbarWidth() {
            return this.f5;
        }
        set verticalViewScrollbarWidth(a) {
            this.f5 = a;
            this.g("VerticalViewScrollbarWidth");
        }
        get verticalViewScrollbarInset() {
            return this.f0;
        }
        set verticalViewScrollbarInset(a) {
            this.f0 = a;
            this.g("VerticalViewScrollbarInset");
        }
        get verticalViewScrollbarTrackStartInset() {
            return this.f4;
        }
        set verticalViewScrollbarTrackStartInset(a) {
            this.f4 = a;
            this.g("VerticalViewScrollbarTrackStartInset");
        }
        get verticalViewScrollbarTrackEndInset() {
            return this.f3;
        }
        set verticalViewScrollbarTrackEndInset(a) {
            this.f3 = a;
            this.g("VerticalViewScrollbarTrackEndInset");
        }
        get verticalViewScrollbarShouldAddAutoTrackInsets() {
            return this.ci;
        }
        set verticalViewScrollbarShouldAddAutoTrackInsets(a) {
            this.ci = a;
            this.g("VerticalViewScrollbarShouldAddAutoTrackInsets");
        }
        get windowRectMinWidth() {
            return this.f7;
        }
        set windowRectMinWidth(a) {
            this.f7 = a;
            this.g("WindowRectMinWidth");
        }
        get windowRectMinHeight() {
            return this.f6;
        }
        set windowRectMinHeight(a) {
            this.f6 = a;
            this.g("WindowRectMinHeight");
        }
        get windowSizeMinWidth() {
            return this.f9;
        }
        set windowSizeMinWidth(a) {
            this.f9 = a;
            this.g("WindowSizeMinWidth");
        }
        get windowSizeMinHeight() {
            return this.f8;
        }
        set windowSizeMinHeight(a) {
            this.f8 = a;
            this.g("WindowSizeMinHeight");
        }
        get seriesAddedRef() {
            return this.m9;
        }
        set seriesAddedRef(a) {
            this.m9 = a;
            this.g("SeriesAddedRef");
        }
        get seriesRemovedRef() {
            return this.ni;
        }
        set seriesRemovedRef(a) {
            this.ni = a;
            this.g("SeriesRemovedRef");
        }
        get seriesPointerEnterRef() {
            return this.ne;
        }
        set seriesPointerEnterRef(a) {
            this.ne = a;
            this.g("SeriesPointerEnterRef");
        }
        get seriesPointerLeaveRef() {
            return this.nf;
        }
        set seriesPointerLeaveRef(a) {
            this.nf = a;
            this.g("SeriesPointerLeaveRef");
        }
        get seriesPointerMoveRef() {
            return this.ng;
        }
        set seriesPointerMoveRef(a) {
            this.ng = a;
            this.g("SeriesPointerMoveRef");
        }
        get seriesPointerDownRef() {
            return this.nd;
        }
        set seriesPointerDownRef(a) {
            this.nd = a;
            this.g("SeriesPointerDownRef");
        }
        get seriesPointerUpRef() {
            return this.nh;
        }
        set seriesPointerUpRef(a) {
            this.nh = a;
            this.g("SeriesPointerUpRef");
        }
        get seriesClickRef() {
            return this.na;
        }
        set seriesClickRef(a) {
            this.na = a;
            this.g("SeriesClickRef");
        }
        get plotAreaPointerEnterRef() {
            return this.m1;
        }
        set plotAreaPointerEnterRef(a) {
            this.m1 = a;
            this.g("PlotAreaPointerEnterRef");
        }
        get plotAreaPointerLeaveRef() {
            return this.m2;
        }
        set plotAreaPointerLeaveRef(a) {
            this.m2 = a;
            this.g("PlotAreaPointerLeaveRef");
        }
        get plotAreaPointerMoveRef() {
            return this.m3;
        }
        set plotAreaPointerMoveRef(a) {
            this.m3 = a;
            this.g("PlotAreaPointerMoveRef");
        }
        get plotAreaPointerDownRef() {
            return this.m0;
        }
        set plotAreaPointerDownRef(a) {
            this.m0 = a;
            this.g("PlotAreaPointerDownRef");
        }
        get plotAreaPointerUpRef() {
            return this.m4;
        }
        set plotAreaPointerUpRef(a) {
            this.m4 = a;
            this.g("PlotAreaPointerUpRef");
        }
        get calloutStyleUpdatingRef() {
            return this.j4;
        }
        set calloutStyleUpdatingRef(a) {
            this.j4 = a;
            this.g("CalloutStyleUpdatingRef");
        }
        get calloutRenderStyleUpdatingRef() {
            return this.jz;
        }
        set calloutRenderStyleUpdatingRef(a) {
            this.jz = a;
            this.g("CalloutRenderStyleUpdatingRef");
        }
        get selectedSeriesItemsChangedRef() {
            return this.m5;
        }
        set selectedSeriesItemsChangedRef(a) {
            this.m5 = a;
            this.g("SelectedSeriesItemsChangedRef");
        }
        get focusedSeriesItemsChangedRef() {
            return this.ma;
        }
        set focusedSeriesItemsChangedRef(a) {
            this.ma = a;
            this.g("FocusedSeriesItemsChangedRef");
        }
        get filterStringErrorsParsingRef() {
            return this.l6;
        }
        set filterStringErrorsParsingRef(a) {
            this.l6 = a;
            this.g("FilterStringErrorsParsingRef");
        }
    }
    DomainChartDescription.$t = /*@__PURE__*/ markType(DomainChartDescription, 'DomainChartDescription', Description.$);
    return DomainChartDescription;
})();
