// igniteui-react-core/esm2015/lib/EllipseVisualData.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PrimitiveVisualData } from "./PrimitiveVisualData";
import { AppearanceHelper } from "./AppearanceHelper";
import { markType } from "./type";
/**
 * @hidden
 */
export let EllipseVisualData = /*@__PURE__*/ (() => {
    class EllipseVisualData extends PrimitiveVisualData {
        get_type() {
            return "Ellipse";
        }
        get type() {
            return this.get_type();
        }
        constructor(a, ..._rest) {
            a = (a == void 0) ? 0 : a;
            switch (a) {
                case 0:
                    {
                        super(1, "ellipse1");
                        this._width = 0;
                        this._height = 0;
                    }
                    break;
                case 1:
                    {
                        let c = _rest[0];
                        let d = _rest[1];
                        super(1, c);
                        this._width = 0;
                        this._height = 0;
                        this.width = d.width;
                        this.height = d.height;
                        AppearanceHelper.p(this.appearance, d);
                    }
                    break;
            }
        }
        get width() {
            return this._width;
        }
        set width(a) {
            this._width = a;
        }
        get height() {
            return this._height;
        }
        set height(a) {
            this._height = a;
        }
        e() {
            return "width: " + this.width + ", height: " + this.height;
        }
        scaleByViewport(a) {
            super.scaleByViewport(a);
            this.width = this.width / a.width;
            this.height = this.height / a.height;
        }
        getPointsOverride(a, b) {
        }
    }
    EllipseVisualData.$t = /*@__PURE__*/ markType(EllipseVisualData, 'EllipseVisualData', PrimitiveVisualData.$);
    return EllipseVisualData;
})();
