// igniteui-react-core/esm2015/lib/EmbeddedRefDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let EmbeddedRefDescription = /*@__PURE__*/ (() => {
    class EmbeddedRefDescription extends Description {
        constructor() {
            super(...arguments);
            this.h = null;
            this.i = null;
        }
        get_type() {
            return "EmbeddedRef";
        }
        get type() {
            return this.get_type();
        }
        get refType() {
            return this.h;
        }
        set refType(a) {
            this.h = a;
            this.g("RefType");
        }
        get value() {
            return this.i;
        }
        set value(a) {
            this.i = a;
            this.g("Value");
        }
    }
    EmbeddedRefDescription.$t = /*@__PURE__*/ markType(EmbeddedRefDescription, 'EmbeddedRefDescription', Description.$);
    return EmbeddedRefDescription;
})();
