// igniteui-react-core/esm2015/lib/ExpansionPanelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ExpansionPanelDescription = /*@__PURE__*/ (() => {
    class ExpansionPanelDescription extends Description {
        get_type() {
            return "ExpansionPanel";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.t = null;
            this.w = null;
            this.o = null;
            this.u = null;
            this.v = null;
            this.x = null;
            this.aa = null;
            this.p = null;
            this.y = null;
            this.z = null;
            this.k = 0;
            this.j = 0;
            this.h = false;
            this.ab = null;
            this.q = null;
            this.ac = null;
            this.ad = null;
            this.s = null;
            this.r = null;
            this.n = null;
            this.ae = null;
            this.af = null;
        }
        get caption() {
            return this.t;
        }
        set caption(a) {
            this.t = a;
            this.g("Caption");
        }
        get captionTextColor() {
            return this.w;
        }
        set captionTextColor(a) {
            this.w = a;
            this.g("CaptionTextColor");
        }
        get actualCaptionTextColor() {
            return this.o;
        }
        set actualCaptionTextColor(a) {
            this.o = a;
            this.g("ActualCaptionTextColor");
        }
        get captionCollapsedTextColor() {
            return this.u;
        }
        set captionCollapsedTextColor(a) {
            this.u = a;
            this.g("CaptionCollapsedTextColor");
        }
        get captionExpandedTextColor() {
            return this.v;
        }
        set captionExpandedTextColor(a) {
            this.v = a;
            this.g("CaptionExpandedTextColor");
        }
        get description() {
            return this.x;
        }
        set description(a) {
            this.x = a;
            this.g("Description");
        }
        get descriptionTextColor() {
            return this.aa;
        }
        set descriptionTextColor(a) {
            this.aa = a;
            this.g("DescriptionTextColor");
        }
        get actualDescriptionTextColor() {
            return this.p;
        }
        set actualDescriptionTextColor(a) {
            this.p = a;
            this.g("ActualDescriptionTextColor");
        }
        get descriptionCollapsedTextColor() {
            return this.y;
        }
        set descriptionCollapsedTextColor(a) {
            this.y = a;
            this.g("DescriptionCollapsedTextColor");
        }
        get descriptionExpandedTextColor() {
            return this.z;
        }
        set descriptionExpandedTextColor(a) {
            this.z = a;
            this.g("DescriptionExpandedTextColor");
        }
        get elevation() {
            return this.k;
        }
        set elevation(a) {
            this.k = a;
            this.g("Elevation");
        }
        get actualElevation() {
            return this.j;
        }
        set actualElevation(a) {
            this.j = a;
            this.g("ActualElevation");
        }
        get expanded() {
            return this.h;
        }
        set expanded(a) {
            this.h = a;
            this.g("Expanded");
        }
        get headerBackgroundColor() {
            return this.ab;
        }
        set headerBackgroundColor(a) {
            this.ab = a;
            this.g("HeaderBackgroundColor");
        }
        get actualHeaderBackgroundColor() {
            return this.q;
        }
        set actualHeaderBackgroundColor(a) {
            this.q = a;
            this.g("ActualHeaderBackgroundColor");
        }
        get headerCollapsedBackgroundColor() {
            return this.ac;
        }
        set headerCollapsedBackgroundColor(a) {
            this.ac = a;
            this.g("HeaderCollapsedBackgroundColor");
        }
        get headerExpandedBackgroundColor() {
            return this.ad;
        }
        set headerExpandedBackgroundColor(a) {
            this.ad = a;
            this.g("HeaderExpandedBackgroundColor");
        }
        get actualUmbraShadowColor() {
            return this.s;
        }
        set actualUmbraShadowColor(a) {
            this.s = a;
            this.g("ActualUmbraShadowColor");
        }
        get actualPenumbraShadowColor() {
            return this.r;
        }
        set actualPenumbraShadowColor(a) {
            this.r = a;
            this.g("ActualPenumbraShadowColor");
        }
        get actualAmbientShadowColor() {
            return this.n;
        }
        set actualAmbientShadowColor(a) {
            this.n = a;
            this.g("ActualAmbientShadowColor");
        }
        get onCollapsedRef() {
            return this.ae;
        }
        set onCollapsedRef(a) {
            this.ae = a;
            this.g("OnCollapsedRef");
        }
        get onExpandedRef() {
            return this.af;
        }
        set onExpandedRef(a) {
            this.af = a;
            this.g("OnExpandedRef");
        }
    }
    ExpansionPanelDescription.$t = /*@__PURE__*/ markType(ExpansionPanelDescription, 'ExpansionPanelDescription', Description.$);
    return ExpansionPanelDescription;
})();
