// igniteui-react-core/esm2015/lib/FilterSyntaxParseResult.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, String_$type, markType } from "./type";
import { List$1 } from "./List$1";
import { stringJoin1 } from "./string";
/**
 * @hidden
 */
export let FilterSyntaxParseResult = /*@__PURE__*/ (() => {
    class FilterSyntaxParseResult extends Base {
        constructor() {
            super();
            this.a = null;
            this.d = 0;
            this.c = null;
            this.c = new List$1(String_$type, 0);
        }
        get b() {
            return this.c.count > 0;
        }
        get e() {
            return stringJoin1(String_$type, "\r\n", this.c);
        }
    }
    FilterSyntaxParseResult.$t = /*@__PURE__*/ markType(FilterSyntaxParseResult, 'FilterSyntaxParseResult');
    return FilterSyntaxParseResult;
})();
