// igniteui-react-core/esm2015/lib/FilterRowDefinitionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let FilterRowDefinitionDescription = /*@__PURE__*/ (() => {
    class FilterRowDefinitionDescription extends DefinitionBaseDescription {
        get_type() {
            return "FilterRowDefinition";
        }
        constructor() {
            super();
            this.cq = 0;
            this.cs = 0;
            this.cr = 0;
            this.cp = 0;
        }
        get paddingLeft() {
            return this.cq;
        }
        set paddingLeft(a) {
            this.cq = a;
            this.g("PaddingLeft");
        }
        get paddingTop() {
            return this.cs;
        }
        set paddingTop(a) {
            this.cs = a;
            this.g("PaddingTop");
        }
        get paddingRight() {
            return this.cr;
        }
        set paddingRight(a) {
            this.cr = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.cp;
        }
        set paddingBottom(a) {
            this.cp = a;
            this.g("PaddingBottom");
        }
    }
    FilterRowDefinitionDescription.$t = /*@__PURE__*/ markType(FilterRowDefinitionDescription, 'FilterRowDefinitionDescription', DefinitionBaseDescription.$);
    FilterRowDefinitionDescription.__canMarshalByValue = true;
    FilterRowDefinitionDescription.__marshalByValueAlias = "FilterRowDefinition";
    return FilterRowDefinitionDescription;
})();
