// igniteui-react-core/esm2015/lib/FinancialCalculationDataSourceDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let FinancialCalculationDataSourceDescription = /*@__PURE__*/ (() => {
    class FinancialCalculationDataSourceDescription extends Description {
        get_type() {
            return "FinancialCalculationDataSource";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.i = null;
            this.h = null;
            this.z = 0;
            this.aa = 0;
            this.y = 0;
            this.x = 0;
            this.w = 0;
            this.v = 0;
            this.r = 0;
            this.q = 0;
            this.p = 0;
            this.n = false;
        }
        get typicalColumn() {
            return this.j;
        }
        set typicalColumn(a) {
            this.j = a;
            this.g("TypicalColumn");
        }
        get trueRange() {
            return this.i;
        }
        set trueRange(a) {
            this.i = a;
            this.g("TrueRange");
        }
        get trueLow() {
            return this.h;
        }
        set trueLow(a) {
            this.h = a;
            this.g("TrueLow");
        }
        get period() {
            return this.z;
        }
        set period(a) {
            this.z = a;
            this.g("Period");
        }
        get shortPeriod() {
            return this.aa;
        }
        set shortPeriod(a) {
            this.aa = a;
            this.g("ShortPeriod");
        }
        get longPeriod() {
            return this.y;
        }
        set longPeriod(a) {
            this.y = a;
            this.g("LongPeriod");
        }
        get count() {
            return this.x;
        }
        set count(a) {
            this.x = a;
            this.g("Count");
        }
        get calculateFrom() {
            return this.w;
        }
        set calculateFrom(a) {
            this.w = a;
            this.g("CalculateFrom");
        }
        get calculateCount() {
            return this.v;
        }
        set calculateCount(a) {
            this.v = a;
            this.g("CalculateCount");
        }
        get multiplier() {
            return this.r;
        }
        set multiplier(a) {
            this.r = a;
            this.g("Multiplier");
        }
        get minimumValue() {
            return this.q;
        }
        set minimumValue(a) {
            this.q = a;
            this.g("MinimumValue");
        }
        get maximumValue() {
            return this.p;
        }
        set maximumValue(a) {
            this.p = a;
            this.g("MaximumValue");
        }
        get specifiesRange() {
            return this.n;
        }
        set specifiesRange(a) {
            this.n = a;
            this.g("SpecifiesRange");
        }
    }
    FinancialCalculationDataSourceDescription.$t = /*@__PURE__*/ markType(FinancialCalculationDataSourceDescription, 'FinancialCalculationDataSourceDescription', Description.$);
    return FinancialCalculationDataSourceDescription;
})();
