// igniteui-react-core/esm2015/lib/FinancialEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let FinancialEventArgsDescription = /*@__PURE__*/ (() => {
    class FinancialEventArgsDescription extends Description {
        get_type() {
            return "FinancialEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.m = 0;
            this.l = 0;
            this.h = null;
            this.j = null;
        }
        get position() {
            return this.m;
        }
        set position(a) {
            this.m = a;
            this.g("Position");
        }
        get count() {
            return this.l;
        }
        set count(a) {
            this.l = a;
            this.g("Count");
        }
        get dataSource() {
            return this.h;
        }
        set dataSource(a) {
            this.h = a;
            this.g("DataSource");
        }
        get supportingCalculations() {
            return this.j;
        }
        set supportingCalculations(a) {
            this.j = a;
            this.g("SupportingCalculations");
        }
    }
    FinancialEventArgsDescription.$t = /*@__PURE__*/ markType(FinancialEventArgsDescription, 'FinancialEventArgsDescription', Description.$);
    return FinancialEventArgsDescription;
})();
