// igniteui-react-core/esm2015/lib/FinancialCalculationSupportingCalculationsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let FinancialCalculationSupportingCalculationsDescription = /*@__PURE__*/ (() => {
    class FinancialCalculationSupportingCalculationsDescription extends Description {
        get_type() {
            return "FinancialCalculationSupportingCalculations";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = null;
            this.k = null;
            this.i = null;
            this.s = null;
            this.q = null;
            this.r = null;
            this.p = null;
            this.x = null;
        }
        get eMA() {
            return this.h;
        }
        set eMA(a) {
            this.h = a;
            this.g("EMA");
        }
        get sMA() {
            return this.j;
        }
        set sMA(a) {
            this.j = a;
            this.g("SMA");
        }
        get sTDEV() {
            return this.k;
        }
        set sTDEV(a) {
            this.k = a;
            this.g("STDEV");
        }
        get movingSum() {
            return this.i;
        }
        set movingSum(a) {
            this.i = a;
            this.g("MovingSum");
        }
        get shortVolumeOscillatorAverage() {
            return this.s;
        }
        set shortVolumeOscillatorAverage(a) {
            this.s = a;
            this.g("ShortVolumeOscillatorAverage");
        }
        get longVolumeOscillatorAverage() {
            return this.q;
        }
        set longVolumeOscillatorAverage(a) {
            this.q = a;
            this.g("LongVolumeOscillatorAverage");
        }
        get shortPriceOscillatorAverage() {
            return this.r;
        }
        set shortPriceOscillatorAverage(a) {
            this.r = a;
            this.g("ShortPriceOscillatorAverage");
        }
        get longPriceOscillatorAverage() {
            return this.p;
        }
        set longPriceOscillatorAverage(a) {
            this.p = a;
            this.g("LongPriceOscillatorAverage");
        }
        get makeSafeRef() {
            return this.x;
        }
        set makeSafeRef(a) {
            this.x = a;
            this.g("MakeSafeRef");
        }
    }
    FinancialCalculationSupportingCalculationsDescription.$t = /*@__PURE__*/ markType(FinancialCalculationSupportingCalculationsDescription, 'FinancialCalculationSupportingCalculationsDescription', Description.$);
    return FinancialCalculationSupportingCalculationsDescription;
})();
