// igniteui-react-core/esm2015/lib/FinancialIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FinancialSeriesDescription } from "./FinancialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let FinancialIndicatorDescription = /*@__PURE__*/ (() => {
    class FinancialIndicatorDescription extends FinancialSeriesDescription {
        get_type() {
            return "FinancialIndicator";
        }
        constructor() {
            super();
            this.ge = null;
            this.f9 = 0;
            this.gg = null;
            this.gf = null;
            this.gd = null;
            this.f7 = 0;
            this.f5 = null;
            this.ga = 0;
        }
        get displayType() {
            return this.ge;
        }
        set displayType(a) {
            this.ge = a;
            this.g("DisplayType");
        }
        get ignoreFirst() {
            return this.f9;
        }
        set ignoreFirst(a) {
            this.f9 = a;
            this.g("IgnoreFirst");
        }
        get trendLineType() {
            return this.gg;
        }
        set trendLineType(a) {
            this.gg = a;
            this.g("TrendLineType");
        }
        get trendLineBrush() {
            return this.gf;
        }
        set trendLineBrush(a) {
            this.gf = a;
            this.g("TrendLineBrush");
        }
        get actualTrendLineBrush() {
            return this.gd;
        }
        set actualTrendLineBrush(a) {
            this.gd = a;
            this.g("ActualTrendLineBrush");
        }
        get trendLineThickness() {
            return this.f7;
        }
        set trendLineThickness(a) {
            this.f7 = a;
            this.g("TrendLineThickness");
        }
        get trendLineDashArray() {
            return this.f5;
        }
        set trendLineDashArray(a) {
            this.f5 = a;
            this.g("TrendLineDashArray");
        }
        get trendLinePeriod() {
            return this.ga;
        }
        set trendLinePeriod(a) {
            this.ga = a;
            this.g("TrendLinePeriod");
        }
    }
    FinancialIndicatorDescription.$t = /*@__PURE__*/ markType(FinancialIndicatorDescription, 'FinancialIndicatorDescription', FinancialSeriesDescription.$);
    return FinancialIndicatorDescription;
})();
