// igniteui-react-core/esm2015/lib/FocusEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let FocusEventArgsDescription = /*@__PURE__*/ (() => {
    class FocusEventArgsDescription extends Description {
        get_type() {
            return "FocusEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
        }
        get targetRef() {
            return this.i;
        }
        set targetRef(a) {
            this.i = a;
            this.g("TargetRef");
        }
        get fromTargetRef() {
            return this.h;
        }
        set fromTargetRef(a) {
            this.h = a;
            this.g("FromTargetRef");
        }
    }
    FocusEventArgsDescription.$t = /*@__PURE__*/ markType(FocusEventArgsDescription, 'FocusEventArgsDescription', Description.$);
    return FocusEventArgsDescription;
})();
