// igniteui-react-core/esm2015/lib/FormatSummaryTextEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let FormatSummaryTextEventArgsDescription = /*@__PURE__*/ (() => {
    class FormatSummaryTextEventArgsDescription extends Description {
        get_type() {
            return "FormatSummaryTextEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
            this.l = null;
            this.j = null;
            this.k = null;
        }
        get field() {
            return this.i;
        }
        set field(a) {
            this.i = a;
            this.g("Field");
        }
        get displayName() {
            return this.h;
        }
        set displayName(a) {
            this.h = a;
            this.g("DisplayName");
        }
        get summaryResultRef() {
            return this.l;
        }
        set summaryResultRef(a) {
            this.l = a;
            this.g("SummaryResultRef");
        }
        get formattedResult() {
            return this.j;
        }
        set formattedResult(a) {
            this.j = a;
            this.g("FormattedResult");
        }
        get formattedText() {
            return this.k;
        }
        set formattedText(a) {
            this.k = a;
            this.g("FormattedText");
        }
    }
    FormatSummaryTextEventArgsDescription.$t = /*@__PURE__*/ markType(FormatSummaryTextEventArgsDescription, 'FormatSummaryTextEventArgsDescription', Description.$);
    return FormatSummaryTextEventArgsDescription;
})();
