// igniteui-react-core/esm2015/lib/FormatGroupTextEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let FormatGroupTextEventArgsDescription = /*@__PURE__*/ (() => {
    class FormatGroupTextEventArgsDescription extends Description {
        get_type() {
            return "FormatGroupTextEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.k = null;
            this.i = null;
            this.h = null;
        }
        get groupName() {
            return this.j;
        }
        set groupName(a) {
            this.j = a;
            this.g("GroupName");
        }
        get groupValueRef() {
            return this.k;
        }
        set groupValueRef(a) {
            this.k = a;
            this.g("GroupValueRef");
        }
        get formattedValue() {
            return this.i;
        }
        set formattedValue(a) {
            this.i = a;
            this.g("FormattedValue");
        }
        get formattedText() {
            return this.h;
        }
        set formattedText(a) {
            this.h = a;
            this.g("FormattedText");
        }
    }
    FormatGroupTextEventArgsDescription.$t = /*@__PURE__*/ markType(FormatGroupTextEventArgsDescription, 'FormatGroupTextEventArgsDescription', Description.$);
    return FormatGroupTextEventArgsDescription;
})();
