// igniteui-react-core/esm2015/lib/FullStochasticOscillatorIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicatorDescription } from "./StrategyBasedIndicatorDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let FullStochasticOscillatorIndicatorDescription = /*@__PURE__*/ (() => {
    class FullStochasticOscillatorIndicatorDescription extends StrategyBasedIndicatorDescription {
        get_type() {
            return "FullStochasticOscillatorIndicator";
        }
        constructor() {
            super();
            this.gl = 0;
            this.gm = 0;
            this.gn = 0;
        }
        get period() {
            return this.gl;
        }
        set period(a) {
            this.gl = a;
            this.g("Period");
        }
        get smoothingPeriod() {
            return this.gm;
        }
        set smoothingPeriod(a) {
            this.gm = a;
            this.g("SmoothingPeriod");
        }
        get triggerPeriod() {
            return this.gn;
        }
        set triggerPeriod(a) {
            this.gn = a;
            this.g("TriggerPeriod");
        }
    }
    FullStochasticOscillatorIndicatorDescription.$t = /*@__PURE__*/ markType(FullStochasticOscillatorIndicatorDescription, 'FullStochasticOscillatorIndicatorDescription', StrategyBasedIndicatorDescription.$);
    return FullStochasticOscillatorIndicatorDescription;
})();
