// igniteui-react-core/esm2015/lib/FunctionFilterExpressionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FilterExpressionDescription } from "./FilterExpressionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let FunctionFilterExpressionDescription = /*@__PURE__*/ (() => {
    class FunctionFilterExpressionDescription extends FilterExpressionDescription {
        get_type() {
            return "FunctionFilterExpression";
        }
        constructor() {
            super();
            this.o = null;
            this.k = null;
        }
        get functionType() {
            return this.o;
        }
        set functionType(a) {
            this.o = a;
            this.g("FunctionType");
        }
        get functionArguments() {
            return this.k;
        }
        set functionArguments(a) {
            this.k = a;
            this.g("FunctionArguments");
        }
    }
    FunctionFilterExpressionDescription.$t = /*@__PURE__*/ markType(FunctionFilterExpressionDescription, 'FunctionFilterExpressionDescription', FilterExpressionDescription.$);
    FunctionFilterExpressionDescription.__marshalByValue1 = true;
    FunctionFilterExpressionDescription.__marshalByValueAlias1 = "FunctionFilterExpression";
    return FunctionFilterExpressionDescription;
})();
