// igniteui-react-core/esm2015/lib/FunnelSliceEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let FunnelSliceEventArgsDescription = /*@__PURE__*/ (() => {
    class FunnelSliceEventArgsDescription extends Description {
        get_type() {
            return "FunnelSliceEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = 0;
            this.j = null;
            this.n = null;
            this.h = null;
        }
        get index() {
            return this.l;
        }
        set index(a) {
            this.l = a;
            this.g("Index");
        }
        get bounds() {
            return this.j;
        }
        set bounds(a) {
            this.j = a;
            this.g("Bounds");
        }
        get itemRef() {
            return this.n;
        }
        set itemRef(a) {
            this.n = a;
            this.g("ItemRef");
        }
        get position() {
            return this.h;
        }
        set position(a) {
            this.h = a;
            this.g("Position");
        }
    }
    FunnelSliceEventArgsDescription.$t = /*@__PURE__*/ markType(FunnelSliceEventArgsDescription, 'FunnelSliceEventArgsDescription', Description.$);
    return FunnelSliceEventArgsDescription;
})();
