// igniteui-react-core/esm2015/lib/GeographicPolylineSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicShapeSeriesBaseDescription } from "./GeographicShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicPolylineSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicPolylineSeriesDescription extends GeographicShapeSeriesBaseDescription {
        get_type() {
            return "GeographicPolylineSeries";
        }
        constructor() {
            super();
            this.f3 = null;
            this.f1 = null;
            this.f2 = null;
            this.fy = 0;
            this.fx = 0;
            this.f4 = null;
        }
        get shapeStyleSelectorRef() {
            return this.f3;
        }
        set shapeStyleSelectorRef(a) {
            this.f3 = a;
            this.g("ShapeStyleSelectorRef");
        }
        get shapeFill() {
            return this.f1;
        }
        set shapeFill(a) {
            this.f1 = a;
            this.g("ShapeFill");
        }
        get shapeStroke() {
            return this.f2;
        }
        set shapeStroke(a) {
            this.f2 = a;
            this.g("ShapeStroke");
        }
        get shapeStrokeThickness() {
            return this.fy;
        }
        set shapeStrokeThickness(a) {
            this.fy = a;
            this.g("ShapeStrokeThickness");
        }
        get shapeOpacity() {
            return this.fx;
        }
        set shapeOpacity(a) {
            this.fx = a;
            this.g("ShapeOpacity");
        }
        get styleShapeRef() {
            return this.f4;
        }
        set styleShapeRef(a) {
            this.f4 = a;
            this.g("StyleShapeRef");
        }
    }
    GeographicPolylineSeriesDescription.$t = /*@__PURE__*/ markType(GeographicPolylineSeriesDescription, 'GeographicPolylineSeriesDescription', GeographicShapeSeriesBaseDescription.$);
    return GeographicPolylineSeriesDescription;
})();
