// igniteui-react-core/esm2015/lib/GeographicHighDensityScatterSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicHighDensityScatterSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicHighDensityScatterSeriesDescription extends GeographicMapSeriesHostDescription {
        get_type() {
            return "GeographicHighDensityScatterSeries";
        }
        constructor() {
            super();
            this.fl = null;
            this.fm = null;
            this.e8 = false;
            this.e7 = false;
            this.fc = 0;
            this.fb = 0;
            this.fk = null;
            this.fj = null;
            this.ff = 0;
            this.fg = 0;
            this.fn = null;
        }
        get latitudeMemberPath() {
            return this.fl;
        }
        set latitudeMemberPath(a) {
            this.fl = a;
            this.g("LatitudeMemberPath");
        }
        get longitudeMemberPath() {
            return this.fm;
        }
        set longitudeMemberPath(a) {
            this.fm = a;
            this.g("LongitudeMemberPath");
        }
        get useBruteForce() {
            return this.e8;
        }
        set useBruteForce(a) {
            this.e8 = a;
            this.g("UseBruteForce");
        }
        get progressiveLoad() {
            return this.e7;
        }
        set progressiveLoad(a) {
            this.e7 = a;
            this.g("ProgressiveLoad");
        }
        get heatMinimum() {
            return this.fc;
        }
        set heatMinimum(a) {
            this.fc = a;
            this.g("HeatMinimum");
        }
        get heatMaximum() {
            return this.fb;
        }
        set heatMaximum(a) {
            this.fb = a;
            this.g("HeatMaximum");
        }
        get heatMinimumColor() {
            return this.fk;
        }
        set heatMinimumColor(a) {
            this.fk = a;
            this.g("HeatMinimumColor");
        }
        get heatMaximumColor() {
            return this.fj;
        }
        set heatMaximumColor(a) {
            this.fj = a;
            this.g("HeatMaximumColor");
        }
        get pointExtent() {
            return this.ff;
        }
        set pointExtent(a) {
            this.ff = a;
            this.g("PointExtent");
        }
        get progressiveStatus() {
            return this.fg;
        }
        set progressiveStatus(a) {
            this.fg = a;
            this.g("ProgressiveStatus");
        }
        get progressiveLoadStatusChangedRef() {
            return this.fn;
        }
        set progressiveLoadStatusChangedRef(a) {
            this.fn = a;
            this.g("ProgressiveLoadStatusChangedRef");
        }
    }
    GeographicHighDensityScatterSeriesDescription.$t = /*@__PURE__*/ markType(GeographicHighDensityScatterSeriesDescription, 'GeographicHighDensityScatterSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicHighDensityScatterSeriesDescription;
})();
