// igniteui-react-core/esm2015/lib/GeographicMarkerSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicMarkerSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicMarkerSeriesDescription extends GeographicMapSeriesHostDescription {
        get_type() {
            return "GeographicMarkerSeries";
        }
        constructor() {
            super();
            this.fj = null;
            this.e7 = 0;
            this.fh = null;
            this.ff = null;
            this.fi = null;
            this.fe = null;
            this.fg = null;
            this.e9 = 0;
            this.fd = null;
            this.fb = null;
            this.fc = null;
        }
        get markerType() {
            return this.fj;
        }
        set markerType(a) {
            this.fj = a;
            this.g("MarkerType");
        }
        get markerThickness() {
            return this.e7;
        }
        set markerThickness(a) {
            this.e7 = a;
            this.g("MarkerThickness");
        }
        get markerOutlineMode() {
            return this.fh;
        }
        set markerOutlineMode(a) {
            this.fh = a;
            this.g("MarkerOutlineMode");
        }
        get markerFillMode() {
            return this.ff;
        }
        set markerFillMode(a) {
            this.ff = a;
            this.g("MarkerFillMode");
        }
        get markerTemplateRef() {
            return this.fi;
        }
        set markerTemplateRef(a) {
            this.fi = a;
            this.g("MarkerTemplateRef");
        }
        get markerBrush() {
            return this.fe;
        }
        set markerBrush(a) {
            this.fe = a;
            this.g("MarkerBrush");
        }
        get markerOutline() {
            return this.fg;
        }
        set markerOutline(a) {
            this.fg = a;
            this.g("MarkerOutline");
        }
        get maximumMarkers() {
            return this.e9;
        }
        set maximumMarkers(a) {
            this.e9 = a;
            this.g("MaximumMarkers");
        }
        get actualMarkerTemplateRef() {
            return this.fd;
        }
        set actualMarkerTemplateRef(a) {
            this.fd = a;
            this.g("ActualMarkerTemplateRef");
        }
        get actualMarkerBrush() {
            return this.fb;
        }
        set actualMarkerBrush(a) {
            this.fb = a;
            this.g("ActualMarkerBrush");
        }
        get actualMarkerOutline() {
            return this.fc;
        }
        set actualMarkerOutline(a) {
            this.fc = a;
            this.g("ActualMarkerOutline");
        }
    }
    GeographicMarkerSeriesDescription.$t = /*@__PURE__*/ markType(GeographicMarkerSeriesDescription, 'GeographicMarkerSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicMarkerSeriesDescription;
})();
