// igniteui-react-core/esm2015/lib/GeographicSymbolSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMarkerSeriesDescription } from "./GeographicMarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicSymbolSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicSymbolSeriesDescription extends GeographicMarkerSeriesDescription {
        get_type() {
            return "GeographicSymbolSeries";
        }
        constructor() {
            super();
            this.f2 = null;
            this.f3 = null;
            this.f4 = null;
            this.fu = false;
            this.ft = false;
            this.f1 = null;
            this.fx = 0;
            this.f0 = null;
            this.fz = null;
        }
        get latitudeMemberPath() {
            return this.f2;
        }
        set latitudeMemberPath(a) {
            this.f2 = a;
            this.g("LatitudeMemberPath");
        }
        get longitudeMemberPath() {
            return this.f3;
        }
        set longitudeMemberPath(a) {
            this.f3 = a;
            this.g("LongitudeMemberPath");
        }
        get markerCollisionAvoidance() {
            return this.f4;
        }
        set markerCollisionAvoidance(a) {
            this.f4 = a;
            this.g("MarkerCollisionAvoidance");
        }
        get isCustomScatterStyleAllowed() {
            return this.fu;
        }
        set isCustomScatterStyleAllowed(a) {
            this.fu = a;
            this.g("IsCustomScatterStyleAllowed");
        }
        get isCustomScatterMarkerStyleAllowed() {
            return this.ft;
        }
        set isCustomScatterMarkerStyleAllowed(a) {
            this.ft = a;
            this.g("IsCustomScatterMarkerStyleAllowed");
        }
        get itemSearchMode() {
            return this.f1;
        }
        set itemSearchMode(a) {
            this.f1 = a;
            this.g("ItemSearchMode");
        }
        get itemSearchThreshold() {
            return this.fx;
        }
        set itemSearchThreshold(a) {
            this.fx = a;
            this.g("ItemSearchThreshold");
        }
        get assigningScatterStyleRef() {
            return this.f0;
        }
        set assigningScatterStyleRef(a) {
            this.f0 = a;
            this.g("AssigningScatterStyleRef");
        }
        get assigningScatterMarkerStyleRef() {
            return this.fz;
        }
        set assigningScatterMarkerStyleRef(a) {
            this.fz = a;
            this.g("AssigningScatterMarkerStyleRef");
        }
    }
    GeographicSymbolSeriesDescription.$t = /*@__PURE__*/ markType(GeographicSymbolSeriesDescription, 'GeographicSymbolSeriesDescription', GeographicMarkerSeriesDescription.$);
    return GeographicSymbolSeriesDescription;
})();
