// igniteui-react-core/esm2015/lib/GeographicContourLineSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicXYTriangulatingSeriesDescription } from "./GeographicXYTriangulatingSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicContourLineSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicContourLineSeriesDescription extends GeographicXYTriangulatingSeriesDescription {
        get_type() {
            return "GeographicContourLineSeries";
        }
        constructor() {
            super();
            this.fq = null;
            this.fm = null;
            this.fl = null;
            this.fj = null;
            this.fp = null;
        }
        get valueMemberPath() {
            return this.fq;
        }
        set valueMemberPath(a) {
            this.fq = a;
            this.g("ValueMemberPath");
        }
        get fillScale() {
            return this.fm;
        }
        set fillScale(a) {
            this.fm = a;
            this.g("FillScale");
        }
        get actualFillScale() {
            return this.fl;
        }
        set actualFillScale(a) {
            this.fl = a;
            this.g("ActualFillScale");
        }
        get valueResolver() {
            return this.fj;
        }
        set valueResolver(a) {
            this.fj = a;
            this.g("ValueResolver");
        }
        get triangulationStatusChangedRef() {
            return this.fp;
        }
        set triangulationStatusChangedRef(a) {
            this.fp = a;
            this.g("TriangulationStatusChangedRef");
        }
    }
    GeographicContourLineSeriesDescription.$t = /*@__PURE__*/ markType(GeographicContourLineSeriesDescription, 'GeographicContourLineSeriesDescription', GeographicXYTriangulatingSeriesDescription.$);
    return GeographicContourLineSeriesDescription;
})();
