// igniteui-react-core/esm2015/lib/GeographicScatterAreaSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicXYTriangulatingSeriesDescription } from "./GeographicXYTriangulatingSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicScatterAreaSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicScatterAreaSeriesDescription extends GeographicXYTriangulatingSeriesDescription {
        get_type() {
            return "GeographicScatterAreaSeries";
        }
        constructor() {
            super();
            this.fn = null;
            this.fk = null;
            this.fj = null;
            this.fo = null;
        }
        get colorMemberPath() {
            return this.fn;
        }
        set colorMemberPath(a) {
            this.fn = a;
            this.g("ColorMemberPath");
        }
        get colorScale() {
            return this.fk;
        }
        set colorScale(a) {
            this.fk = a;
            this.g("ColorScale");
        }
        get actualColorScale() {
            return this.fj;
        }
        set actualColorScale(a) {
            this.fj = a;
            this.g("ActualColorScale");
        }
        get triangulationStatusChangedRef() {
            return this.fo;
        }
        set triangulationStatusChangedRef(a) {
            this.fo = a;
            this.g("TriangulationStatusChangedRef");
        }
    }
    GeographicScatterAreaSeriesDescription.$t = /*@__PURE__*/ markType(GeographicScatterAreaSeriesDescription, 'GeographicScatterAreaSeriesDescription', GeographicXYTriangulatingSeriesDescription.$);
    return GeographicScatterAreaSeriesDescription;
})();
