// igniteui-react-core/esm2015/lib/GeographicMapSeriesHostDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicMapSeriesHostDescription = /*@__PURE__*/ (() => {
    class GeographicMapSeriesHostDescription extends SeriesDescription {
        get_type() {
            return "GeographicMapSeriesHost";
        }
        constructor() {
            super();
            this.e3 = 0;
            this.e5 = null;
        }
        get visibleFromScale() {
            return this.e3;
        }
        set visibleFromScale(a) {
            this.e3 = a;
            this.g("VisibleFromScale");
        }
        get coercionMethodsRef() {
            return this.e5;
        }
        set coercionMethodsRef(a) {
            this.e5 = a;
            this.g("CoercionMethodsRef");
        }
    }
    GeographicMapSeriesHostDescription.$t = /*@__PURE__*/ markType(GeographicMapSeriesHostDescription, 'GeographicMapSeriesHostDescription', SeriesDescription.$);
    return GeographicMapSeriesHostDescription;
})();
