// igniteui-react-core/esm2015/lib/GetTileImageUriArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GetTileImageUriArgsDescription = /*@__PURE__*/ (() => {
    class GetTileImageUriArgsDescription extends Description {
        get_type() {
            return "GetTileImageUriArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = 0;
            this.i = 0;
            this.j = 0;
            this.n = null;
        }
        get tileLevel() {
            return this.h;
        }
        set tileLevel(a) {
            this.h = a;
            this.g("TileLevel");
        }
        get tilePositionX() {
            return this.i;
        }
        set tilePositionX(a) {
            this.i = a;
            this.g("TilePositionX");
        }
        get tilePositionY() {
            return this.j;
        }
        set tilePositionY(a) {
            this.j = a;
            this.g("TilePositionY");
        }
        get tileImageUri() {
            return this.n;
        }
        set tileImageUri(a) {
            this.n = a;
            this.g("TileImageUri");
        }
    }
    GetTileImageUriArgsDescription.$t = /*@__PURE__*/ markType(GetTileImageUriArgsDescription, 'GetTileImageUriArgsDescription', Description.$);
    return GetTileImageUriArgsDescription;
})();
