// igniteui-react-core/esm2015/lib/GridSelectedKeysChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridSelectedKeysChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridSelectedKeysChangedEventArgsDescription extends Description {
        get_type() {
            return "GridSelectedKeysChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
            this.j = null;
        }
        get currentKeys() {
            return this.i;
        }
        set currentKeys(a) {
            this.i = a;
            this.g("CurrentKeys");
        }
        get addedKeys() {
            return this.h;
        }
        set addedKeys(a) {
            this.h = a;
            this.g("AddedKeys");
        }
        get removedKeys() {
            return this.j;
        }
        set removedKeys(a) {
            this.j = a;
            this.g("RemovedKeys");
        }
    }
    GridSelectedKeysChangedEventArgsDescription.$t = /*@__PURE__*/ markType(GridSelectedKeysChangedEventArgsDescription, 'GridSelectedKeysChangedEventArgsDescription', Description.$);
    return GridSelectedKeysChangedEventArgsDescription;
})();
