// igniteui-react-core/esm2015/lib/GridFilterDialogViewModelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridFilterDialogViewModelDescription = /*@__PURE__*/ (() => {
    class GridFilterDialogViewModelDescription extends Description {
        get_type() {
            return "GridFilterDialogViewModel";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.j = false;
            this.l = 0;
            this.h = null;
        }
        get propertyType() {
            return this.n;
        }
        set propertyType(a) {
            this.n = a;
            this.g("PropertyType");
        }
        get isTopLevelOr() {
            return this.j;
        }
        set isTopLevelOr(a) {
            this.j = a;
            this.g("IsTopLevelOr");
        }
        get maxGroupingLevels() {
            return this.l;
        }
        set maxGroupingLevels(a) {
            this.l = a;
            this.g("MaxGroupingLevels");
        }
        get rows() {
            return this.h;
        }
        set rows(a) {
            this.h = a;
            this.g("Rows");
        }
    }
    GridFilterDialogViewModelDescription.$t = /*@__PURE__*/ markType(GridFilterDialogViewModelDescription, 'GridFilterDialogViewModelDescription', Description.$);
    return GridFilterDialogViewModelDescription;
})();
