// igniteui-react-core/esm2015/lib/GridColumnPinOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GridColumnOptionsSimpleSectionBaseDescription } from "./GridColumnOptionsSimpleSectionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridColumnPinOptionsDescription = /*@__PURE__*/ (() => {
    class GridColumnPinOptionsDescription extends GridColumnOptionsSimpleSectionBaseDescription {
        get_type() {
            return "GridColumnPinOptions";
        }
        constructor() {
            super();
            this.aw = null;
            this.ax = null;
        }
        get pinLeftCaption() {
            return this.aw;
        }
        set pinLeftCaption(a) {
            this.aw = a;
            this.g("PinLeftCaption");
        }
        get pinRightCaption() {
            return this.ax;
        }
        set pinRightCaption(a) {
            this.ax = a;
            this.g("PinRightCaption");
        }
    }
    GridColumnPinOptionsDescription.$t = /*@__PURE__*/ markType(GridColumnPinOptionsDescription, 'GridColumnPinOptionsDescription', GridColumnOptionsSimpleSectionBaseDescription.$);
    return GridColumnPinOptionsDescription;
})();
