// igniteui-react-core/esm2015/lib/GridColumnButtonOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GridColumnOptionsSectionBaseDescription } from "./GridColumnOptionsSectionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridColumnButtonOptionsDescription = /*@__PURE__*/ (() => {
    class GridColumnButtonOptionsDescription extends GridColumnOptionsSectionBaseDescription {
        get_type() {
            return "GridColumnButtonOptions";
        }
        constructor() {
            super();
            this.as = null;
            this.au = null;
            this.at = null;
            this.av = null;
        }
        get applyButtonCaption() {
            return this.as;
        }
        set applyButtonCaption(a) {
            this.as = a;
            this.g("ApplyButtonCaption");
        }
        get cancelButtonCaption() {
            return this.au;
        }
        set cancelButtonCaption(a) {
            this.au = a;
            this.g("CancelButtonCaption");
        }
        get applyButtonClickRef() {
            return this.at;
        }
        set applyButtonClickRef(a) {
            this.at = a;
            this.g("ApplyButtonClickRef");
        }
        get cancelButtonClickRef() {
            return this.av;
        }
        set cancelButtonClickRef(a) {
            this.av = a;
            this.g("CancelButtonClickRef");
        }
    }
    GridColumnButtonOptionsDescription.$t = /*@__PURE__*/ markType(GridColumnButtonOptionsDescription, 'GridColumnButtonOptionsDescription', GridColumnOptionsSectionBaseDescription.$);
    return GridColumnButtonOptionsDescription;
})();
