// igniteui-react-core/esm2015/lib/GridColumnSummaryOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GridColumnOptionsSectionBaseDescription } from "./GridColumnOptionsSectionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridColumnSummaryOptionsDescription = /*@__PURE__*/ (() => {
    class GridColumnSummaryOptionsDescription extends GridColumnOptionsSectionBaseDescription {
        get_type() {
            return "GridColumnSummaryOptions";
        }
        constructor() {
            super();
            this.av = null;
            this.ax = null;
            this.au = null;
            this.as = 0;
            this.ay = null;
            this.aw = null;
        }
        get summaryCaption() {
            return this.av;
        }
        set summaryCaption(a) {
            this.av = a;
            this.g("SummaryCaption");
        }
        get summaryListDensity() {
            return this.ax;
        }
        set summaryListDensity(a) {
            this.ax = a;
            this.g("SummaryListDensity");
        }
        get actualSummaryListDensity() {
            return this.au;
        }
        set actualSummaryListDensity(a) {
            this.au = a;
            this.g("ActualSummaryListDensity");
        }
        get activeCount() {
            return this.as;
        }
        set activeCount(a) {
            this.as = a;
            this.g("ActiveCount");
        }
        get summaryListTextColor() {
            return this.ay;
        }
        set summaryListTextColor(a) {
            this.ay = a;
            this.g("SummaryListTextColor");
        }
        get summaryListBackground() {
            return this.aw;
        }
        set summaryListBackground(a) {
            this.aw = a;
            this.g("SummaryListBackground");
        }
    }
    GridColumnSummaryOptionsDescription.$t = /*@__PURE__*/ markType(GridColumnSummaryOptionsDescription, 'GridColumnSummaryOptionsDescription', GridColumnOptionsSectionBaseDescription.$);
    return GridColumnSummaryOptionsDescription;
})();
