// igniteui-react-core/esm2015/lib/GridBridgeColumnDefinition.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { isNaN_ } from "./number";
/**
 * @hidden
 */
export let GridBridgeColumnDefinition = /*@__PURE__*/ (() => {
    class GridBridgeColumnDefinition extends Base {
        constructor() {
            super(...arguments);
            this.a = 0;
            this.d = 0;
            this.b = NaN;
        }
        get c() {
            return this.b;
        }
        set c(a) {
            this.b = a;
        }
        e() {
            switch (this.a) {
                case 0: return "auto";
                case 1: return this.d + "px";
                case 2: return isNaN_(this.c) ? this.d + "fr" : "minmax(" + this.c + "px," + this.d + "fr)";
            }
            return "";
        }
    }
    GridBridgeColumnDefinition.$t = /*@__PURE__*/ markType(GridBridgeColumnDefinition, 'GridBridgeColumnDefinition');
    return GridBridgeColumnDefinition;
})();
