// igniteui-react-core/esm2015/lib/GridColumnMoveOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GridColumnOptionsSimpleSectionBaseDescription } from "./GridColumnOptionsSimpleSectionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridColumnMoveOptionsDescription = /*@__PURE__*/ (() => {
    class GridColumnMoveOptionsDescription extends GridColumnOptionsSimpleSectionBaseDescription {
        get_type() {
            return "GridColumnMoveOptions";
        }
        constructor() {
            super();
            this.aw = null;
            this.ax = null;
        }
        get moveLeftCaption() {
            return this.aw;
        }
        set moveLeftCaption(a) {
            this.aw = a;
            this.g("MoveLeftCaption");
        }
        get moveRightCaption() {
            return this.ax;
        }
        set moveRightCaption(a) {
            this.ax = a;
            this.g("MoveRightCaption");
        }
    }
    GridColumnMoveOptionsDescription.$t = /*@__PURE__*/ markType(GridColumnMoveOptionsDescription, 'GridColumnMoveOptionsDescription', GridColumnOptionsSimpleSectionBaseDescription.$);
    return GridColumnMoveOptionsDescription;
})();
