// igniteui-react-core/esm2015/lib/GridFilterDialogViewModelRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridFilterDialogViewModelRowDescription = /*@__PURE__*/ (() => {
    class GridFilterDialogViewModelRowDescription extends Description {
        get_type() {
            return "GridFilterDialogViewModelRow";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.l = null;
            this.h = null;
            this.m = null;
            this.n = null;
        }
        get operators() {
            return this.i;
        }
        set operators(a) {
            this.i = a;
            this.g("Operators");
        }
        get currentOperator() {
            return this.l;
        }
        set currentOperator(a) {
            this.l = a;
            this.g("CurrentOperator");
        }
        get operandNumber() {
            return this.h;
        }
        set operandNumber(a) {
            this.h = a;
            this.g("OperandNumber");
        }
        get operand1Ref() {
            return this.m;
        }
        set operand1Ref(a) {
            this.m = a;
            this.g("Operand1Ref");
        }
        get operand2Ref() {
            return this.n;
        }
        set operand2Ref(a) {
            this.n = a;
            this.g("Operand2Ref");
        }
    }
    GridFilterDialogViewModelRowDescription.$t = /*@__PURE__*/ markType(GridFilterDialogViewModelRowDescription, 'GridFilterDialogViewModelRowDescription', Description.$);
    return GridFilterDialogViewModelRowDescription;
})();
