// igniteui-react-core/esm2015/lib/GridDataCommittedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridDataCommittedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridDataCommittedEventArgsDescription extends Description {
        get_type() {
            return "GridDataCommittedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = 0;
            this.h = null;
        }
        get commitID() {
            return this.j;
        }
        set commitID(a) {
            this.j = a;
            this.g("CommitID");
        }
        get changes() {
            return this.h;
        }
        set changes(a) {
            this.h = a;
            this.g("Changes");
        }
    }
    GridDataCommittedEventArgsDescription.$t = /*@__PURE__*/ markType(GridDataCommittedEventArgsDescription, 'GridDataCommittedEventArgsDescription', Description.$);
    return GridDataCommittedEventArgsDescription;
})();
