// igniteui-react-core/esm2015/lib/GridCellValueChangingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridCellValueChangingEventArgsDescription = /*@__PURE__*/ (() => {
    class GridCellValueChangingEventArgsDescription extends Description {
        get_type() {
            return "GridCellValueChangingEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = 0;
            this.h = null;
            this.j = null;
            this.r = null;
            this.o = null;
            this.n = null;
        }
        get editID() {
            return this.l;
        }
        set editID(a) {
            this.l = a;
            this.g("EditID");
        }
        get cellInfo() {
            return this.h;
        }
        set cellInfo(a) {
            this.h = a;
            this.g("CellInfo");
        }
        get column() {
            return this.j;
        }
        set column(a) {
            this.j = a;
            this.g("Column");
        }
        get itemRef() {
            return this.r;
        }
        set itemRef(a) {
            this.r = a;
            this.g("ItemRef");
        }
        get oldValue() {
            return this.o;
        }
        set oldValue(a) {
            this.o = a;
            this.g("OldValue");
        }
        get newValue() {
            return this.n;
        }
        set newValue(a) {
            this.n = a;
            this.g("NewValue");
        }
    }
    GridCellValueChangingEventArgsDescription.$t = /*@__PURE__*/ markType(GridCellValueChangingEventArgsDescription, 'GridCellValueChangingEventArgsDescription', Description.$);
    return GridCellValueChangingEventArgsDescription;
})();
