// igniteui-react-core/esm2015/lib/GridColumnFilterOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GridColumnOptionsSectionBaseDescription } from "./GridColumnOptionsSectionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridColumnFilterOptionsDescription = /*@__PURE__*/ (() => {
    class GridColumnFilterOptionsDescription extends GridColumnOptionsSectionBaseDescription {
        get_type() {
            return "GridColumnFilterOptions";
        }
        constructor() {
            super();
            this.at = null;
            this.av = null;
            this.au = null;
            this.as = null;
            this.aw = null;
        }
        get clearColumnFiltersCaption() {
            return this.at;
        }
        set clearColumnFiltersCaption(a) {
            this.at = a;
            this.g("ClearColumnFiltersCaption");
        }
        get filterListPlaceholderText() {
            return this.av;
        }
        set filterListPlaceholderText(a) {
            this.av = a;
            this.g("FilterListPlaceholderText");
        }
        get filterListDensity() {
            return this.au;
        }
        set filterListDensity(a) {
            this.au = a;
            this.g("FilterListDensity");
        }
        get actualFilterListDensity() {
            return this.as;
        }
        set actualFilterListDensity(a) {
            this.as = a;
            this.g("ActualFilterListDensity");
        }
        get textFilterCaption() {
            return this.aw;
        }
        set textFilterCaption(a) {
            this.aw = a;
            this.g("TextFilterCaption");
        }
    }
    GridColumnFilterOptionsDescription.$t = /*@__PURE__*/ markType(GridColumnFilterOptionsDescription, 'GridColumnFilterOptionsDescription', GridColumnOptionsSectionBaseDescription.$);
    return GridColumnFilterOptionsDescription;
})();
