// igniteui-react-core/esm2015/lib/GridAnimationPhaseSettingsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridAnimationPhaseSettingsDescription = /*@__PURE__*/ (() => {
    class GridAnimationPhaseSettingsDescription extends Description {
        get_type() {
            return "GridAnimationPhaseSettings";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = 0;
            this.l = 0;
            this.m = 0;
            this.n = 0;
            this.j = 0;
            this.h = false;
            this.t = null;
        }
        get durationMilliseconds() {
            return this.k;
        }
        set durationMilliseconds(a) {
            this.k = a;
            this.g("DurationMilliseconds");
        }
        get holdInitialMilliseconds() {
            return this.l;
        }
        set holdInitialMilliseconds(a) {
            this.l = a;
            this.g("HoldInitialMilliseconds");
        }
        get perItemDelayMilliseconds() {
            return this.m;
        }
        set perItemDelayMilliseconds(a) {
            this.m = a;
            this.g("PerItemDelayMilliseconds");
        }
        get subItemDurationMilliseconds() {
            return this.n;
        }
        set subItemDurationMilliseconds(a) {
            this.n = a;
            this.g("SubItemDurationMilliseconds");
        }
        get desiredSubItemDurationMilliseconds() {
            return this.j;
        }
        set desiredSubItemDurationMilliseconds(a) {
            this.j = a;
            this.g("DesiredSubItemDurationMilliseconds");
        }
        get shouldItemsFinishSimultaneously() {
            return this.h;
        }
        set shouldItemsFinishSimultaneously(a) {
            this.h = a;
            this.g("ShouldItemsFinishSimultaneously");
        }
        get easingFunctionType() {
            return this.t;
        }
        set easingFunctionType(a) {
            this.t = a;
            this.g("EasingFunctionType");
        }
    }
    GridAnimationPhaseSettingsDescription.$t = /*@__PURE__*/ markType(GridAnimationPhaseSettingsDescription, 'GridAnimationPhaseSettingsDescription', Description.$);
    return GridAnimationPhaseSettingsDescription;
})();
