// igniteui-react-core/esm2015/lib/GridSelectedCellRangesChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridSelectedCellRangesChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridSelectedCellRangesChangedEventArgsDescription extends Description {
        get_type() {
            return "GridSelectedCellRangesChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
            this.j = null;
            this.k = null;
        }
        get currentRanges() {
            return this.i;
        }
        set currentRanges(a) {
            this.i = a;
            this.g("CurrentRanges");
        }
        get addedRanges() {
            return this.h;
        }
        set addedRanges(a) {
            this.h = a;
            this.g("AddedRanges");
        }
        get removedRanges() {
            return this.j;
        }
        set removedRanges(a) {
            this.j = a;
            this.g("RemovedRanges");
        }
        get updatedRanges() {
            return this.k;
        }
        set updatedRanges(a) {
            this.k = a;
            this.g("UpdatedRanges");
        }
    }
    GridSelectedCellRangesChangedEventArgsDescription.$t = /*@__PURE__*/ markType(GridSelectedCellRangesChangedEventArgsDescription, 'GridSelectedCellRangesChangedEventArgsDescription', Description.$);
    return GridSelectedCellRangesChangedEventArgsDescription;
})();
