// igniteui-react-core/esm2015/lib/GridSizeChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridSizeChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridSizeChangedEventArgsDescription extends Description {
        get_type() {
            return "GridSizeChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = 0;
            this.h = 0;
        }
        get width() {
            return this.i;
        }
        set width(a) {
            this.i = a;
            this.g("Width");
        }
        get height() {
            return this.h;
        }
        set height(a) {
            this.h = a;
            this.g("Height");
        }
    }
    GridSizeChangedEventArgsDescription.$t = /*@__PURE__*/ markType(GridSizeChangedEventArgsDescription, 'GridSizeChangedEventArgsDescription', Description.$);
    return GridSizeChangedEventArgsDescription;
})();
