// igniteui-react-core/esm2015/lib/GridColumnOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GridColumnOptionsBaseDescription } from "./GridColumnOptionsBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridColumnOptionsDescription = /*@__PURE__*/ (() => {
    class GridColumnOptionsDescription extends GridColumnOptionsBaseDescription {
        get_type() {
            return "GridColumnOptions";
        }
        constructor() {
            super();
            this.bp = null;
            this.a8 = 0;
            this.bq = null;
            this.br = null;
            this.bb = 0;
            this.bd = 0;
            this.bc = 0;
            this.ba = 0;
            this.au = false;
            this.ay = false;
            this.aw = false;
            this.ax = false;
            this.av = false;
            this.at = false;
            this.as = false;
            this.az = false;
            this.b2 = null;
            this.b0 = null;
            this.b1 = null;
            this.bu = null;
            this.bv = null;
            this.bw = null;
            this.bx = null;
            this.by = null;
            this.bz = null;
            this.bl = null;
            this.bn = null;
            this.bo = null;
            this.b4 = null;
            this.bj = null;
            this.bs = null;
            this.bi = null;
            this.bt = null;
            this.b5 = null;
            this.b3 = null;
            this.bk = null;
            this.bm = null;
        }
        get columnNameFontFamily() {
            return this.bp;
        }
        set columnNameFontFamily(a) {
            this.bp = a;
            this.g("ColumnNameFontFamily");
        }
        get columnNameFontSize() {
            return this.a8;
        }
        set columnNameFontSize(a) {
            this.a8 = a;
            this.g("ColumnNameFontSize");
        }
        get columnNameFontStyle() {
            return this.bq;
        }
        set columnNameFontStyle(a) {
            this.bq = a;
            this.g("ColumnNameFontStyle");
        }
        get columnNameFontWeight() {
            return this.br;
        }
        set columnNameFontWeight(a) {
            this.br = a;
            this.g("ColumnNameFontWeight");
        }
        get paddingLeft() {
            return this.bb;
        }
        set paddingLeft(a) {
            this.bb = a;
            this.g("PaddingLeft");
        }
        get paddingTop() {
            return this.bd;
        }
        set paddingTop(a) {
            this.bd = a;
            this.g("PaddingTop");
        }
        get paddingRight() {
            return this.bc;
        }
        set paddingRight(a) {
            this.bc = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.ba;
        }
        set paddingBottom(a) {
            this.ba = a;
            this.g("PaddingBottom");
        }
        get headerVisible() {
            return this.au;
        }
        set headerVisible(a) {
            this.au = a;
            this.g("HeaderVisible");
        }
        get sortOptionsVisible() {
            return this.ay;
        }
        set sortOptionsVisible(a) {
            this.ay = a;
            this.g("SortOptionsVisible");
        }
        get moveOptionsVisible() {
            return this.aw;
        }
        set moveOptionsVisible(a) {
            this.aw = a;
            this.g("MoveOptionsVisible");
        }
        get pinOptionsVisible() {
            return this.ax;
        }
        set pinOptionsVisible(a) {
            this.ax = a;
            this.g("PinOptionsVisible");
        }
        get hideOptionsVisible() {
            return this.av;
        }
        set hideOptionsVisible(a) {
            this.av = a;
            this.g("HideOptionsVisible");
        }
        get groupOptionsVisible() {
            return this.at;
        }
        set groupOptionsVisible(a) {
            this.at = a;
            this.g("GroupOptionsVisible");
        }
        get filterOptionsVisible() {
            return this.as;
        }
        set filterOptionsVisible(a) {
            this.as = a;
            this.g("FilterOptionsVisible");
        }
        get summaryOptionsVisible() {
            return this.az;
        }
        set summaryOptionsVisible(a) {
            this.az = a;
            this.g("SummaryOptionsVisible");
        }
        get sortHeaderCaption() {
            return this.b2;
        }
        set sortHeaderCaption(a) {
            this.b2 = a;
            this.g("SortHeaderCaption");
        }
        get sortAscendingCaption() {
            return this.b0;
        }
        set sortAscendingCaption(a) {
            this.b0 = a;
            this.g("SortAscendingCaption");
        }
        get sortDescendingCaption() {
            return this.b1;
        }
        set sortDescendingCaption(a) {
            this.b1 = a;
            this.g("SortDescendingCaption");
        }
        get moveHeaderCaption() {
            return this.bu;
        }
        set moveHeaderCaption(a) {
            this.bu = a;
            this.g("MoveHeaderCaption");
        }
        get moveLeftCaption() {
            return this.bv;
        }
        set moveLeftCaption(a) {
            this.bv = a;
            this.g("MoveLeftCaption");
        }
        get moveRightCaption() {
            return this.bw;
        }
        set moveRightCaption(a) {
            this.bw = a;
            this.g("MoveRightCaption");
        }
        get pinHeaderCaption() {
            return this.bx;
        }
        set pinHeaderCaption(a) {
            this.bx = a;
            this.g("PinHeaderCaption");
        }
        get pinLeftCaption() {
            return this.by;
        }
        set pinLeftCaption(a) {
            this.by = a;
            this.g("PinLeftCaption");
        }
        get pinRightCaption() {
            return this.bz;
        }
        set pinRightCaption(a) {
            this.bz = a;
            this.g("PinRightCaption");
        }
        get applyFiltersButtonCaption() {
            return this.bl;
        }
        set applyFiltersButtonCaption(a) {
            this.bl = a;
            this.g("ApplyFiltersButtonCaption");
        }
        get cancelFiltersButtonCaption() {
            return this.bn;
        }
        set cancelFiltersButtonCaption(a) {
            this.bn = a;
            this.g("CancelFiltersButtonCaption");
        }
        get clearColumnFiltersCaption() {
            return this.bo;
        }
        set clearColumnFiltersCaption(a) {
            this.bo = a;
            this.g("ClearColumnFiltersCaption");
        }
        get summaryListDensity() {
            return this.b4;
        }
        set summaryListDensity(a) {
            this.b4 = a;
            this.g("SummaryListDensity");
        }
        get actualSummaryListDensity() {
            return this.bj;
        }
        set actualSummaryListDensity(a) {
            this.bj = a;
            this.g("ActualSummaryListDensity");
        }
        get filterListDensity() {
            return this.bs;
        }
        set filterListDensity(a) {
            this.bs = a;
            this.g("FilterListDensity");
        }
        get actualFilterListDensity() {
            return this.bi;
        }
        set actualFilterListDensity(a) {
            this.bi = a;
            this.g("ActualFilterListDensity");
        }
        get filterListPlaceholderText() {
            return this.bt;
        }
        set filterListPlaceholderText(a) {
            this.bt = a;
            this.g("FilterListPlaceholderText");
        }
        get summaryListTextColor() {
            return this.b5;
        }
        set summaryListTextColor(a) {
            this.b5 = a;
            this.g("SummaryListTextColor");
        }
        get summaryListBackground() {
            return this.b3;
        }
        set summaryListBackground(a) {
            this.b3 = a;
            this.g("SummaryListBackground");
        }
        get applyButtonClickRef() {
            return this.bk;
        }
        set applyButtonClickRef(a) {
            this.bk = a;
            this.g("ApplyButtonClickRef");
        }
        get cancelButtonClickRef() {
            return this.bm;
        }
        set cancelButtonClickRef(a) {
            this.bm = a;
            this.g("CancelButtonClickRef");
        }
    }
    GridColumnOptionsDescription.$t = /*@__PURE__*/ markType(GridColumnOptionsDescription, 'GridColumnOptionsDescription', GridColumnOptionsBaseDescription.$);
    return GridColumnOptionsDescription;
})();
