// igniteui-react-core/esm2015/lib/GridCellPositionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridCellPositionDescription = /*@__PURE__*/ (() => {
    class GridCellPositionDescription extends Description {
        get_type() {
            return "GridCellPosition";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = 0;
            this.l = null;
        }
        get rowIndex() {
            return this.i;
        }
        set rowIndex(a) {
            this.i = a;
            this.g("RowIndex");
        }
        get columnUniqueKey() {
            return this.l;
        }
        set columnUniqueKey(a) {
            this.l = a;
            this.g("ColumnUniqueKey");
        }
    }
    GridCellPositionDescription.$t = /*@__PURE__*/ markType(GridCellPositionDescription, 'GridCellPositionDescription', Description.$);
    GridCellPositionDescription.__marshalByValue = true;
    GridCellPositionDescription.__marshalByValueAlias = "GridCellPosition";
    return GridCellPositionDescription;
})();
