// igniteui-react-core/esm2015/lib/GridColumnSortOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GridColumnOptionsSimpleSectionBaseDescription } from "./GridColumnOptionsSimpleSectionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridColumnSortOptionsDescription = /*@__PURE__*/ (() => {
    class GridColumnSortOptionsDescription extends GridColumnOptionsSimpleSectionBaseDescription {
        get_type() {
            return "GridColumnSortOptions";
        }
        constructor() {
            super();
            this.aw = null;
            this.ax = null;
            this.ay = null;
        }
        get ascendingCaption() {
            return this.aw;
        }
        set ascendingCaption(a) {
            this.aw = a;
            this.g("AscendingCaption");
        }
        get descendingCaption() {
            return this.ax;
        }
        set descendingCaption(a) {
            this.ax = a;
            this.g("DescendingCaption");
        }
        get sortDirection() {
            return this.ay;
        }
        set sortDirection(a) {
            this.ay = a;
            this.g("SortDirection");
        }
    }
    GridColumnSortOptionsDescription.$t = /*@__PURE__*/ markType(GridColumnSortOptionsDescription, 'GridColumnSortOptionsDescription', GridColumnOptionsSimpleSectionBaseDescription.$);
    return GridColumnSortOptionsDescription;
})();
