// igniteui-react-core/esm2015/lib/GridColumnWidthChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridColumnWidthChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridColumnWidthChangedEventArgsDescription extends Description {
        get_type() {
            return "GridColumnWidthChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.i = null;
            this.h = null;
        }
        get column() {
            return this.l;
        }
        set column(a) {
            this.l = a;
            this.g("Column");
        }
        get oldWidth() {
            return this.i;
        }
        set oldWidth(a) {
            this.i = a;
            this.g("OldWidth");
        }
        get newWidth() {
            return this.h;
        }
        set newWidth(a) {
            this.h = a;
            this.g("NewWidth");
        }
    }
    GridColumnWidthChangedEventArgsDescription.$t = /*@__PURE__*/ markType(GridColumnWidthChangedEventArgsDescription, 'GridColumnWidthChangedEventArgsDescription', Description.$);
    return GridColumnWidthChangedEventArgsDescription;
})();
