// igniteui-react-core/esm2015/lib/HeaderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let HeaderDescription = /*@__PURE__*/ (() => {
    class HeaderDescription extends DefinitionBaseDescription {
        get_type() {
            return "Header";
        }
        constructor() {
            super();
            this.c0 = null;
            this.cy = null;
            this.c1 = null;
            this.cz = null;
            this.cq = 0;
            this.cs = 0;
            this.cr = 0;
            this.cp = 0;
        }
        get sortIndicatorColor() {
            return this.c0;
        }
        set sortIndicatorColor(a) {
            this.c0 = a;
            this.g("SortIndicatorColor");
        }
        get actualSortIndicatorColor() {
            return this.cy;
        }
        set actualSortIndicatorColor(a) {
            this.cy = a;
            this.g("ActualSortIndicatorColor");
        }
        get sortIndicatorStyle() {
            return this.c1;
        }
        set sortIndicatorStyle(a) {
            this.c1 = a;
            this.g("SortIndicatorStyle");
        }
        get actualSortIndicatorStyle() {
            return this.cz;
        }
        set actualSortIndicatorStyle(a) {
            this.cz = a;
            this.g("ActualSortIndicatorStyle");
        }
        get paddingLeft() {
            return this.cq;
        }
        set paddingLeft(a) {
            this.cq = a;
            this.g("PaddingLeft");
        }
        get paddingTop() {
            return this.cs;
        }
        set paddingTop(a) {
            this.cs = a;
            this.g("PaddingTop");
        }
        get paddingRight() {
            return this.cr;
        }
        set paddingRight(a) {
            this.cr = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.cp;
        }
        set paddingBottom(a) {
            this.cp = a;
            this.g("PaddingBottom");
        }
    }
    HeaderDescription.$t = /*@__PURE__*/ markType(HeaderDescription, 'HeaderDescription', DefinitionBaseDescription.$);
    HeaderDescription.__canMarshalByValue = true;
    HeaderDescription.__marshalByValueAlias = "Header";
    return HeaderDescription;
})();
