// igniteui-react-core/esm2015/lib/HeatTileGeneratorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { TileGeneratorDescription } from "./TileGeneratorDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let HeatTileGeneratorDescription = /*@__PURE__*/ (() => {
    class HeatTileGeneratorDescription extends TileGeneratorDescription {
        get_type() {
            return "HeatTileGenerator";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.m = null;
            this.l = null;
            this.k = null;
            this.an = null;
            this.am = null;
            this.j = null;
            this.i = null;
            this.ac = 0;
            this.ae = 0;
            this.s = false;
            this.t = false;
            this.u = false;
            this.ag = 0;
            this.af = 0;
            this.ad = 0;
            this.v = false;
            this.w = false;
            this.ap = null;
            this.ao = null;
        }
        get yValues() {
            return this.m;
        }
        set yValues(a) {
            this.m = a;
            this.g("YValues");
        }
        get xValues() {
            return this.l;
        }
        set xValues(a) {
            this.l = a;
            this.g("XValues");
        }
        get values() {
            return this.k;
        }
        set values(a) {
            this.k = a;
            this.g("Values");
        }
        get minimumColor() {
            return this.an;
        }
        set minimumColor(a) {
            this.an = a;
            this.g("MinimumColor");
        }
        get maximumColor() {
            return this.am;
        }
        set maximumColor(a) {
            this.am = a;
            this.g("MaximumColor");
        }
        get scaleColors() {
            return this.j;
        }
        set scaleColors(a) {
            this.j = a;
            this.g("ScaleColors");
        }
        get scaleColorOffsets() {
            return this.i;
        }
        set scaleColorOffsets(a) {
            this.i = a;
            this.g("ScaleColorOffsets");
        }
        get blurRadius() {
            return this.ac;
        }
        set blurRadius(a) {
            this.ac = a;
            this.g("BlurRadius");
        }
        get maxBlurRadius() {
            return this.ae;
        }
        set maxBlurRadius(a) {
            this.ae = a;
            this.g("MaxBlurRadius");
        }
        get useBlurRadiusAdjustedForZoom() {
            return this.s;
        }
        set useBlurRadiusAdjustedForZoom(a) {
            this.s = a;
            this.g("UseBlurRadiusAdjustedForZoom");
        }
        get useGlobalMinMax() {
            return this.t;
        }
        set useGlobalMinMax(a) {
            this.t = a;
            this.g("UseGlobalMinMax");
        }
        get useGlobalMinMaxAdjustedForZoom() {
            return this.u;
        }
        set useGlobalMinMaxAdjustedForZoom(a) {
            this.u = a;
            this.g("UseGlobalMinMaxAdjustedForZoom");
        }
        get minimumValue() {
            return this.ag;
        }
        set minimumValue(a) {
            this.ag = a;
            this.g("MinimumValue");
        }
        get maximumValue() {
            return this.af;
        }
        set maximumValue(a) {
            this.af = a;
            this.g("MaximumValue");
        }
        get logarithmBase() {
            return this.ad;
        }
        set logarithmBase(a) {
            this.ad = a;
            this.g("LogarithmBase");
        }
        get useLogarithmicScale() {
            return this.v;
        }
        set useLogarithmicScale(a) {
            this.v = a;
            this.g("UseLogarithmicScale");
        }
        get useWebWorkers() {
            return this.w;
        }
        set useWebWorkers(a) {
            this.w = a;
            this.g("UseWebWorkers");
        }
        get webWorkerScriptPath() {
            return this.ap;
        }
        set webWorkerScriptPath(a) {
            this.ap = a;
            this.g("WebWorkerScriptPath");
        }
        get webWorkerInstanceRef() {
            return this.ao;
        }
        set webWorkerInstanceRef(a) {
            this.ao = a;
            this.g("WebWorkerInstanceRef");
        }
    }
    HeatTileGeneratorDescription.$t = /*@__PURE__*/ markType(HeatTileGeneratorDescription, 'HeatTileGeneratorDescription', TileGeneratorDescription.$);
    return HeatTileGeneratorDescription;
})();
