// igniteui-react-core/esm2015/lib/HighlightState.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Enum specifying state of the highlight value
 */
export var HighlightState = /*@__PURE__*/ (function (HighlightState) {
    HighlightState[HighlightState["Hidden"] = 0] = "Hidden";
    HighlightState[HighlightState["StartShowing"] = 1] = "StartShowing";
    HighlightState[HighlightState["Showing"] = 2] = "Showing";
    HighlightState[HighlightState["Shown"] = 3] = "Shown";
    HighlightState[HighlightState["StartHiding"] = 4] = "StartHiding";
    HighlightState[HighlightState["Hiding"] = 5] = "Hiding";
    return HighlightState;
})({});
/**
 * @hidden
 */
export let HighlightState_$type = /*@__PURE__*/ markEnum('HighlightState', 'Hidden,0|StartShowing,1|Showing,2|Shown,3|StartHiding,4|Hiding,5');
