// igniteui-react-core/esm2015/lib/HighlightingInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { HighlightingState } from "./HighlightingState";
/**
 * @hidden
 */
export let HighlightingInfo = /*@__PURE__*/ (() => {
    class HighlightingInfo extends Base {
        constructor() {
            super();
            this.h = 0;
            this.g = 0;
            this.f = 0;
            this.a = 0;
            this.e = 0;
            this.d = false;
            this.c = false;
            this.b = false;
            this.i = null;
            this.a = HighlightingState.outward;
            this.e = 0;
            this.d = false;
        }
    }
    HighlightingInfo.$t = /*@__PURE__*/ markType(HighlightingInfo, 'HighlightingInfo');
    return HighlightingInfo;
})();
