// igniteui-react-core/esm2015/lib/HighDensityScatterSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let HighDensityScatterSeriesDescription = /*@__PURE__*/ (() => {
    class HighDensityScatterSeriesDescription extends SeriesDescription {
        get_type() {
            return "HighDensityScatterSeries";
        }
        constructor() {
            super();
            this.fi = null;
            this.fm = null;
            this.fl = null;
            this.fj = null;
            this.fn = null;
            this.fk = null;
            this.fo = null;
            this.fp = null;
            this.e4 = false;
            this.e3 = false;
            this.e8 = 0;
            this.e7 = 0;
            this.fg = null;
            this.ff = null;
            this.fb = 0;
            this.fc = 0;
            this.fh = null;
        }
        get xAxisRef() {
            return this.fi;
        }
        set xAxisRef(a) {
            this.fi = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.fm;
        }
        set yAxisRef(a) {
            this.fm = a;
            this.g("YAxisRef");
        }
        get xMemberPath() {
            return this.fl;
        }
        set xMemberPath(a) {
            this.fl = a;
            this.g("XMemberPath");
        }
        get xMemberAsLegendLabel() {
            return this.fj;
        }
        set xMemberAsLegendLabel(a) {
            this.fj = a;
            this.g("XMemberAsLegendLabel");
        }
        get yMemberAsLegendLabel() {
            return this.fn;
        }
        set yMemberAsLegendLabel(a) {
            this.fn = a;
            this.g("YMemberAsLegendLabel");
        }
        get xMemberAsLegendUnit() {
            return this.fk;
        }
        set xMemberAsLegendUnit(a) {
            this.fk = a;
            this.g("XMemberAsLegendUnit");
        }
        get yMemberAsLegendUnit() {
            return this.fo;
        }
        set yMemberAsLegendUnit(a) {
            this.fo = a;
            this.g("YMemberAsLegendUnit");
        }
        get yMemberPath() {
            return this.fp;
        }
        set yMemberPath(a) {
            this.fp = a;
            this.g("YMemberPath");
        }
        get useBruteForce() {
            return this.e4;
        }
        set useBruteForce(a) {
            this.e4 = a;
            this.g("UseBruteForce");
        }
        get progressiveLoad() {
            return this.e3;
        }
        set progressiveLoad(a) {
            this.e3 = a;
            this.g("ProgressiveLoad");
        }
        get heatMinimum() {
            return this.e8;
        }
        set heatMinimum(a) {
            this.e8 = a;
            this.g("HeatMinimum");
        }
        get heatMaximum() {
            return this.e7;
        }
        set heatMaximum(a) {
            this.e7 = a;
            this.g("HeatMaximum");
        }
        get heatMinimumColor() {
            return this.fg;
        }
        set heatMinimumColor(a) {
            this.fg = a;
            this.g("HeatMinimumColor");
        }
        get heatMaximumColor() {
            return this.ff;
        }
        set heatMaximumColor(a) {
            this.ff = a;
            this.g("HeatMaximumColor");
        }
        get pointExtent() {
            return this.fb;
        }
        set pointExtent(a) {
            this.fb = a;
            this.g("PointExtent");
        }
        get progressiveStatus() {
            return this.fc;
        }
        set progressiveStatus(a) {
            this.fc = a;
            this.g("ProgressiveStatus");
        }
        get progressiveLoadStatusChangedRef() {
            return this.fh;
        }
        set progressiveLoadStatusChangedRef(a) {
            this.fh = a;
            this.g("ProgressiveLoadStatusChangedRef");
        }
    }
    HighDensityScatterSeriesDescription.$t = /*@__PURE__*/ markType(HighDensityScatterSeriesDescription, 'HighDensityScatterSeriesDescription', SeriesDescription.$);
    return HighDensityScatterSeriesDescription;
})();
