// igniteui-react-core/esm2015/lib/HighlightedValueLabelMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Describes the label mode for highlighted series.
 */
export var HighlightedValueLabelMode = /*@__PURE__*/ (function (HighlightedValueLabelMode) {
    HighlightedValueLabelMode[HighlightedValueLabelMode["Auto"] = 0] = "Auto";
    HighlightedValueLabelMode[HighlightedValueLabelMode["PreferOriginal"] = 1] = "PreferOriginal";
    HighlightedValueLabelMode[HighlightedValueLabelMode["PreferHighlighted"] = 2] = "PreferHighlighted";
    HighlightedValueLabelMode[HighlightedValueLabelMode["LabelBoth"] = 3] = "LabelBoth";
    return HighlightedValueLabelMode;
})({});
/**
 * @hidden
 */
export let HighlightedValueLabelMode_$type = /*@__PURE__*/ markEnum('HighlightedValueLabelMode', 'Auto,0|PreferOriginal,1|PreferHighlighted,2|LabelBoth,3');
