// igniteui-react-core/esm2015/lib/HorizontalRangeCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RangeCategorySeriesDescription } from "./RangeCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let HorizontalRangeCategorySeriesDescription = /*@__PURE__*/ (() => {
    class HorizontalRangeCategorySeriesDescription extends RangeCategorySeriesDescription {
        get_type() {
            return "HorizontalRangeCategorySeries";
        }
        constructor() {
            super();
            this.gj = null;
            this.gk = null;
            this.gh = null;
            this.gf = null;
            this.gi = null;
            this.gg = null;
        }
        get xAxisRef() {
            return this.gj;
        }
        set xAxisRef(a) {
            this.gj = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.gk;
        }
        set yAxisRef(a) {
            this.gk = a;
            this.g("YAxisRef");
        }
        get lowMemberAsLegendLabel() {
            return this.gh;
        }
        set lowMemberAsLegendLabel(a) {
            this.gh = a;
            this.g("LowMemberAsLegendLabel");
        }
        get highMemberAsLegendLabel() {
            return this.gf;
        }
        set highMemberAsLegendLabel(a) {
            this.gf = a;
            this.g("HighMemberAsLegendLabel");
        }
        get lowMemberAsLegendUnit() {
            return this.gi;
        }
        set lowMemberAsLegendUnit(a) {
            this.gi = a;
            this.g("LowMemberAsLegendUnit");
        }
        get highMemberAsLegendUnit() {
            return this.gg;
        }
        set highMemberAsLegendUnit(a) {
            this.gg = a;
            this.g("HighMemberAsLegendUnit");
        }
    }
    HorizontalRangeCategorySeriesDescription.$t = /*@__PURE__*/ markType(HorizontalRangeCategorySeriesDescription, 'HorizontalRangeCategorySeriesDescription', RangeCategorySeriesDescription.$);
    return HorizontalRangeCategorySeriesDescription;
})();
