// igniteui-react-core/esm2015/lib/IgcCheckboxBridge.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIElementBridge } from "./NativeUIElementBridge";
import { runOn, markType } from "./type";
import { NativeUICheckboxChangeEventArgs } from "./NativeUICheckboxChangeEventArgs";
/**
 * @hidden
 */
export let IgcCheckboxBridge = /*@__PURE__*/ (() => {
    class IgcCheckboxBridge extends NativeUIElementBridge {
        constructor() {
            super(...arguments);
            this.t = null;
            this.s = null;
            this.u = null;
        }
        addHandler(a, b, c, d) {
            switch (c) {
                case 2:
                    let e = a;
                    this.t = e.listen("igcChange", runOn(this, this.v));
                    this.s = d;
                    this.u = b;
                    break;
            }
        }
        v(a) {
            if (this.s != null) {
                let b = new NativeUICheckboxChangeEventArgs();
                this.s(this.u, b);
            }
        }
        getValue(a, b) {
            switch (b) {
                case 4: return true;
                case 7: return a.getProperty("checked");
            }
            return null;
        }
        removeHandler(a, b, c, d) {
            switch (c) {
                case 2:
                    if (this.t != null) {
                        this.t();
                    }
                    this.s = null;
                    this.u = null;
                    break;
            }
        }
        setValue(a, b, c) {
            switch (b) {
                case 4: return;
                case 7:
                    a.setProperty("checked", c);
                    return;
            }
        }
    }
    IgcCheckboxBridge.$t = /*@__PURE__*/ markType(IgcCheckboxBridge, 'IgcCheckboxBridge', NativeUIElementBridge.$);
    return IgcCheckboxBridge;
})();
