// igniteui-react-core/esm2015/lib/igr-data-legend-summary-column.js
import { IgrDataLegendSeriesValueInfo } from "./igr-data-legend-series-value-info";
import { DataLegendSummaryColumn as DataLegendSummaryColumn_internal } from "./DataLegendSummaryColumn";
/**
 * Represent a column in summary row
*/
export class IgrDataLegendSummaryColumn extends IgrDataLegendSeriesValueInfo {
    createImplementation() {
        return new DataLegendSummaryColumn_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    /**
     * Get values of series in the summary column
    */
    get seriesValues() {
        return this.i.ac;
    }
    /**
     * Get labels rendered on left side of series values, e.g. "H:" for High, "L:" for Low
    */
    get seriesLabels() {
        return this.i.aa;
    }
    /**
     * Get optional units on right side of series values, e.g. "Hz" for hertz
    */
    get seriesUnits() {
        return this.i.ab;
    }
    /**
     * Add a value in the summary column
    
    */
    addValue(value) {
        this.i.ak(value);
    }
    /**
     * Add label in the summary column
    
    */
    addLabel(label) {
        this.i.ai(label);
    }
    /**
     * Add unit in the summary column
    
    */
    addUnits(units) {
        this.i.aj(units);
    }
}
