// igniteui-react-core/esm2015/lib/igr-date-time-format-specifier.js
import { IgrFormatSpecifier } from "./igr-format-specifier";
import { DateTimeFormatSpecifier as DateTimeFormatSpecifier_internal } from "./DateTimeFormatSpecifier";
export class IgrDateTimeFormatSpecifier extends IgrFormatSpecifier {
    createImplementation() {
        return new DateTimeFormatSpecifier_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    get locale() {
        return this.i.locale;
    }
    set locale(v) {
        this.i.locale = v;
    }
    get dateStyle() {
        return this.i.dateStyle;
    }
    set dateStyle(v) {
        this.i.dateStyle = v;
    }
    get timeStyle() {
        return this.i.timeStyle;
    }
    set timeStyle(v) {
        this.i.timeStyle = v;
    }
    get fractionalSecondDigits() {
        return this.i.fractionalSecondDigits;
    }
    set fractionalSecondDigits(v) {
        this.i.fractionalSecondDigits = +v;
    }
    get calendar() {
        return this.i.calendar;
    }
    set calendar(v) {
        this.i.calendar = v;
    }
    get dayPeriod() {
        return this.i.dayPeriod;
    }
    set dayPeriod(v) {
        this.i.dayPeriod = v;
    }
    get numberingSystem() {
        return this.i.numberingSystem;
    }
    set numberingSystem(v) {
        this.i.numberingSystem = v;
    }
    get localeMatcher() {
        return this.i.localeMatcher;
    }
    set localeMatcher(v) {
        this.i.localeMatcher = v;
    }
    get timeZone() {
        return this.i.timeZone;
    }
    set timeZone(v) {
        this.i.timeZone = v;
    }
    get hour12() {
        return this.i.hour12;
    }
    set hour12(v) {
        this.i.hour12 = v;
    }
    get hourCycle() {
        return this.i.hourCycle;
    }
    set hourCycle(v) {
        this.i.hourCycle = v;
    }
    get formatMatcher() {
        return this.i.formatMatcher;
    }
    set formatMatcher(v) {
        this.i.formatMatcher = v;
    }
    get weekDay() {
        return this.i.weekDay;
    }
    set weekDay(v) {
        this.i.weekDay = v;
    }
    get era() {
        return this.i.era;
    }
    set era(v) {
        this.i.era = v;
    }
    get year() {
        return this.i.year;
    }
    set year(v) {
        this.i.year = v;
    }
    get month() {
        return this.i.month;
    }
    set month(v) {
        this.i.month = v;
    }
    get day() {
        return this.i.day;
    }
    set day(v) {
        this.i.day = v;
    }
    get hour() {
        return this.i.hour;
    }
    set hour(v) {
        this.i.hour = v;
    }
    get minute() {
        return this.i.minute;
    }
    set minute(v) {
        this.i.minute = v;
    }
    get second() {
        return this.i.second;
    }
    set second(v) {
        this.i.second = v;
    }
    get timeZoneName() {
        return this.i.timeZoneName;
    }
    set timeZoneName(v) {
        this.i.timeZoneName = v;
    }
}
