// igniteui-react-core/esm2015/lib/igr-cancel-event-args.js
import { CancelEventArgs as CancelEventArgs_internal } from "./CancelEventArgs";
import { ensureBool } from "./componentUtil";
export class IgrCancelEventArgs {
    createImplementation() {
        return new CancelEventArgs_internal(0);
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get cancel() {
        return this.i.cancel;
    }
    set cancel(v) {
        this.i.cancel = ensureBool(v);
    }
}
