// igniteui-react-core/esm2015/lib/igr-number-format-specifier.js
import { IgrFormatSpecifier } from "./igr-format-specifier";
import { NumberFormatSpecifier as NumberFormatSpecifier_internal } from "./NumberFormatSpecifier";
import { ensureBool } from "./componentUtil";
export class IgrNumberFormatSpecifier extends IgrFormatSpecifier {
    createImplementation() {
        return new NumberFormatSpecifier_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    get locale() {
        return this.i.locale;
    }
    set locale(v) {
        this.i.locale = v;
    }
    get compactDisplay() {
        return this.i.compactDisplay;
    }
    set compactDisplay(v) {
        this.i.compactDisplay = v;
    }
    get currency() {
        return this.i.currency;
    }
    set currency(v) {
        this.i.currency = v;
    }
    get currencyDisplay() {
        return this.i.currencyDisplay;
    }
    set currencyDisplay(v) {
        this.i.currencyDisplay = v;
    }
    get currencySign() {
        return this.i.currencySign;
    }
    set currencySign(v) {
        this.i.currencySign = v;
    }
    get currencyCode() {
        return this.i.currencyCode;
    }
    set currencyCode(v) {
        this.i.currencyCode = v;
    }
    get localeMatcher() {
        return this.i.localeMatcher;
    }
    set localeMatcher(v) {
        this.i.localeMatcher = v;
    }
    get notation() {
        return this.i.notation;
    }
    set notation(v) {
        this.i.notation = v;
    }
    get numberingSystem() {
        return this.i.numberingSystem;
    }
    set numberingSystem(v) {
        this.i.numberingSystem = v;
    }
    get signDisplay() {
        return this.i.signDisplay;
    }
    set signDisplay(v) {
        this.i.signDisplay = v;
    }
    get style() {
        return this.i.style;
    }
    set style(v) {
        this.i.style = v;
    }
    get unit() {
        return this.i.unit;
    }
    set unit(v) {
        this.i.unit = v;
    }
    get unitDisplay() {
        return this.i.unitDisplay;
    }
    set unitDisplay(v) {
        this.i.unitDisplay = v;
    }
    get useGrouping() {
        return this.i.useGrouping;
    }
    set useGrouping(v) {
        this.i.useGrouping = ensureBool(v);
    }
    get minimumIntegerDigits() {
        return this.i.minimumIntegerDigits;
    }
    set minimumIntegerDigits(v) {
        this.i.minimumIntegerDigits = +v;
    }
    get minimumFractionDigits() {
        return this.i.minimumFractionDigits;
    }
    set minimumFractionDigits(v) {
        this.i.minimumFractionDigits = +v;
    }
    get maximumFractionDigits() {
        return this.i.maximumFractionDigits;
    }
    set maximumFractionDigits(v) {
        this.i.maximumFractionDigits = +v;
    }
    get minimumSignificantDigits() {
        return this.i.minimumSignificantDigits;
    }
    set minimumSignificantDigits(v) {
        this.i.minimumSignificantDigits = +v;
    }
    get maximumSignificantDigits() {
        return this.i.maximumSignificantDigits;
    }
    set maximumSignificantDigits(v) {
        this.i.maximumSignificantDigits = +v;
    }
}
