// igniteui-react-core/esm2015/lib/igr-cancelling-multi-scale-image-event-args.js
import { CancellingMultiScaleImageEventArgs as CancellingMultiScaleImageEventArgs_internal } from "./CancellingMultiScaleImageEventArgs";
export class IgrCancellingMultiScaleImageEventArgs {
    createImplementation() {
        return new CancellingMultiScaleImageEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get uri() {
        return this.i.uri;
    }
    set uri(v) {
        this.i.uri = v;
    }
}
