// igniteui-react-core/esm2015/lib/igr-data-source-row-expansion-changed-event-args.js
/**
 * Provides information about row expansion or collapse.
*/
export class IgrDataSourceRowExpansionChangedEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * The index of the row being expanded or collapsed.
    */
    get rowIndex() {
        return this.i.rowIndex;
    }
    /**
     * The previous state of the row. True is expanded, False is collapsed.
    */
    get oldState() {
        return this.i.oldState;
    }
    /**
     * The new state of the row. True is expanded, False is collapsed.
    */
    get newState() {
        return this.i.newState;
    }
}
