// igniteui-react-core/esm2015/lib/igr-property-updated-event-args.js
/**
 * EventArgs class for property updated events.
*/
export class IgrPropertyUpdatedEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * Name of the property being updated.
    */
    get propertyName() {
        return this.i.propertyName;
    }
    set propertyName(v) {
        this.i.propertyName = v;
    }
    /**
     * Old or previous value of the property being updated.
    */
    get oldValue() {
        return this.i.oldValue;
    }
    set oldValue(v) {
        this.i.oldValue = v;
    }
    /**
     * New or current value of the property being updated.
    */
    get newValue() {
        return this.i.newValue;
    }
    set newValue(v) {
        this.i.newValue = v;
    }
}
