// igniteui-react-core/esm2015/lib/igr-focus-event-args.js
import { FocusEventArgs as FocusEventArgs_internal } from "./FocusEventArgs";
export class IgrFocusEventArgs {
    createImplementation() {
        return new FocusEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get target() {
        return this.i.b;
    }
    set target(v) {
        this.i.b = v;
    }
    get fromTarget() {
        return this.i.a;
    }
    set fromTarget(v) {
        this.i.a = v;
    }
}
