// igniteui-react-core/esm2015/lib/igr-key-event-args.js
/**
 * Event args for KeyUp and KeyDown events.
*/
export class IgrKeyEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get originalEvent() {
        return this.i.originalEvent;
    }
    /**
     * Gets a value indicating whether the ALT key was pressed.
    */
    get alt() {
        return this.i.alt;
    }
    /**
     * Gets a value indicating whether the CTRL key was pressed.
    */
    get ctrl() {
        return this.i.ctrl;
    }
    /**
     * Gets the keyboard code for a KeyDown or KeyUp event.
    */
    get keyCode() {
        return this.i.keyCode;
    }
    /**
     * Gets the modifier flags for a KeyDown or KeyUp event.
     * The flags indicate which combination of CTRL, SHIFT, and ALT keys was pressed.
    */
    get modifiers() {
        return this.i.modifiers;
    }
    /**
     * Gets a value indicating whether the SHIFT key was pressed.
    */
    get shift() {
        return this.i.shift;
    }
    /**
     * Gets whether the PreventDefault method was called.
    */
    get defaultPrevented() {
        return this.i.defaultPrevented;
    }
    /**
     * Tells the user agent that if the event does not get explicitly handled, its default action should not be taken as it normally would be.
    
    */
    preventDefault() {
        this.i.preventDefault();
    }
    /**
     * Prevent the event from bubbling up.
    
    */
    stopPropagation() {
        this.i.stopPropagation();
    }
}
