// igniteui-react-core/esm2015/lib/igr-image-captured-event-args.js
import { IgrCaptureImageSettings } from "./igr-capture-image-settings";
export class IgrImageCapturedEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * Gets the captured image as Base64.
    */
    get base64Data() {
        return this.i.base64Data;
    }
    /**
     * Gets the captured image.
    */
    get image() {
        return this.i.image;
    }
    /**
     * Gets the settings used to capture the image.
    */
    get settings() {
        const r = this.i.settings;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrCaptureImageSettings();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
}
