// igniteui-react-core/esm2015/lib/igr-rect-changed-event-args.js
import { fromRect, toRect } from "./componentUtil";
/**
 * Provides data for rectangle changed events.
*/
export class IgrRectChangedEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * Gets the rectangle before the change.
    */
    get oldRect() {
        return fromRect(this.i.oldRect);
    }
    set oldRect(v) {
        this.i.oldRect = toRect(v);
    }
    /**
     * Gets the new rectangle.
    */
    get newRect() {
        return fromRect(this.i.newRect);
    }
    set newRect(v) {
        this.i.newRect = toRect(v);
    }
}
