// igniteui-react-core/esm2015/lib/igr-stock-changed-event-args.js
/**
 * Represents event arguments for FDC3 MessageRecived
*/
export class IgrStockChangedEventArgs {
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * Gets array of added stock symbols
    */
    get addedSymbols() {
        return this.i.addedSymbols;
    }
    set addedSymbols(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.addedSymbols = v;
    }
    /**
     * Gets array of removed stock symbols
    */
    get removedSymbols() {
        return this.i.removedSymbols;
    }
    set removedSymbols(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.removedSymbols = v;
    }
}
