// igniteui-react-core/esm2015/lib/igr-data-source-root-summaries-changed-event-args.js
import { DataSourceRootSummariesChangedEventArgs as DataSourceRootSummariesChangedEventArgs_internal } from "./DataSourceRootSummariesChangedEventArgs";
/**
 * Provides information about row expansion or collapse.
*/
export class IgrDataSourceRootSummariesChangedEventArgs {
    createImplementation() {
        return new DataSourceRootSummariesChangedEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
}
