// igniteui-react-core/esm2015/lib/ImageCapturedEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { EventArgs, markType } from "./type";
/**
 * @hidden
 */
export let ImageCapturedEventArgs = /*@__PURE__*/ (() => {
    class ImageCapturedEventArgs extends EventArgs {
        constructor(a, b, c) {
            super();
            this.f = null;
            this.d = null;
            this.b = null;
            this.f = a;
            this.d = b;
            this.b = c;
        }
        get base64Data() {
            return this.f;
        }
        get image() {
            return this.d;
        }
        get settings() {
            return this.b;
        }
    }
    ImageCapturedEventArgs.$t = /*@__PURE__*/ markType(ImageCapturedEventArgs, 'ImageCapturedEventArgs', EventArgs.$);
    return ImageCapturedEventArgs;
})();
