// igniteui-react-core/esm2015/lib/ImageColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DataGridColumnDescription } from "./DataGridColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ImageColumnDescription = /*@__PURE__*/ (() => {
    class ImageColumnDescription extends DataGridColumnDescription {
        get_type() {
            return "ImageColumn";
        }
        constructor() {
            super();
            this.f9 = null;
            this.f7 = null;
            this.f5 = null;
            this.f1 = null;
            this.f6 = null;
            this.f8 = null;
        }
        get imageStretchOption() {
            return this.f9;
        }
        set imageStretchOption(a) {
            this.f9 = a;
            this.g("ImageStretchOption");
        }
        get editorType() {
            return this.f7;
        }
        set editorType(a) {
            this.f7 = a;
            this.g("EditorType");
        }
        get editorDataSourceRef() {
            return this.f5;
        }
        set editorDataSourceRef(a) {
            this.f5 = a;
            this.g("EditorDataSourceRef");
        }
        get editorDataSource() {
            return this.f1;
        }
        set editorDataSource(a) {
            this.f1 = a;
            this.g("EditorDataSource");
        }
        get editorTextField() {
            return this.f6;
        }
        set editorTextField(a) {
            this.f6 = a;
            this.g("EditorTextField");
        }
        get editorValueField() {
            return this.f8;
        }
        set editorValueField(a) {
            this.f8 = a;
            this.g("EditorValueField");
        }
    }
    ImageColumnDescription.$t = /*@__PURE__*/ markType(ImageColumnDescription, 'ImageColumnDescription', DataGridColumnDescription.$);
    ImageColumnDescription.__canMarshalByValue1 = true;
    ImageColumnDescription.__marshalByValueAlias1 = "ImageColumn";
    return ImageColumnDescription;
})();
