// igniteui-react-core/esm2015/lib/InputChangeEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let InputChangeEventArgsDescription = /*@__PURE__*/ (() => {
    class InputChangeEventArgsDescription extends Description {
        get_type() {
            return "InputChangeEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = false;
        }
        get value() {
            return this.j;
        }
        set value(a) {
            this.j = a;
            this.g("Value");
        }
        get isCompositionInProgress() {
            return this.h;
        }
        set isCompositionInProgress(a) {
            this.h = a;
            this.g("IsCompositionInProgress");
        }
    }
    InputChangeEventArgsDescription.$t = /*@__PURE__*/ markType(InputChangeEventArgsDescription, 'InputChangeEventArgsDescription', Description.$);
    return InputChangeEventArgsDescription;
})();
