// igniteui-react-core/esm2015/lib/ItemLegendDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { LegendBaseDescription } from "./LegendBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ItemLegendDescription = /*@__PURE__*/ (() => {
    class ItemLegendDescription extends LegendBaseDescription {
        get width() {
            return this.y;
        }
        set width(a) {
            this.y = a;
            this.g("Width");
        }
        get height() {
            return this.u;
        }
        set height(a) {
            this.u = a;
            this.g("Height");
        }
        get_type() {
            return "ItemLegend";
        }
        constructor() {
            super();
            this.y = null;
            this.u = null;
            this.v = null;
            this.w = null;
            this.x = null;
        }
        get orientation() {
            return this.v;
        }
        set orientation(a) {
            this.v = a;
            this.g("Orientation");
        }
        get textColor() {
            return this.w;
        }
        set textColor(a) {
            this.w = a;
            this.g("TextColor");
        }
        get textStyle() {
            return this.x;
        }
        set textStyle(a) {
            this.x = a;
            this.g("TextStyle");
        }
    }
    ItemLegendDescription.$t = /*@__PURE__*/ markType(ItemLegendDescription, 'ItemLegendDescription', LegendBaseDescription.$);
    return ItemLegendDescription;
})();
