// igniteui-react-core/esm2015/lib/ItemToolTipLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ItemToolTipLayerDescription = /*@__PURE__*/ (() => {
    class ItemToolTipLayerDescription extends AnnotationLayerDescription {
        get_type() {
            return "ItemToolTipLayer";
        }
        constructor() {
            super();
            this.gd = null;
            this.ge = null;
            this.ga = false;
            this.f9 = false;
        }
        get targetSeriesName() {
            return this.gd;
        }
        set targetSeriesName(a) {
            this.gd = a;
            this.g("TargetSeriesName");
        }
        get targetSeriesRef() {
            return this.ge;
        }
        set targetSeriesRef(a) {
            this.ge = a;
            this.g("TargetSeriesRef");
        }
        get useInterpolation() {
            return this.ga;
        }
        set useInterpolation(a) {
            this.ga = a;
            this.g("UseInterpolation");
        }
        get skipUnknownValues() {
            return this.f9;
        }
        set skipUnknownValues(a) {
            this.f9 = a;
            this.g("SkipUnknownValues");
        }
    }
    ItemToolTipLayerDescription.$t = /*@__PURE__*/ markType(ItemToolTipLayerDescription, 'ItemToolTipLayerDescription', AnnotationLayerDescription.$);
    return ItemToolTipLayerDescription;
})();
