// igniteui-react-core/esm2015/lib/JsonWriter.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { StringBuilder } from "./StringBuilder";
/**
 * @hidden
 */
export let JsonWriter = /*@__PURE__*/ (() => {
    class JsonWriter extends Base {
        constructor() {
            super(...arguments);
            this.d = new StringBuilder(0);
            this.b = 0;
            this.a = true;
        }
        g() {
            this.b++;
        }
        e() {
            this.b--;
        }
        i(a) {
            this.f();
            this.d.l(a);
        }
        f() {
            if (this.a) {
                this.a = false;
                for (let a = 0; a < this.b; a++) {
                    this.d.l("\t");
                }
            }
        }
        j(a) {
            this.f();
            this.d.u(a);
            this.a = true;
        }
        h() {
            this.f();
            this.d.t();
            this.a = true;
        }
        toString() {
            return this.d.toString();
        }
    }
    JsonWriter.$t = /*@__PURE__*/ markType(JsonWriter, 'JsonWriter');
    return JsonWriter;
})();
