// igniteui-react-core/esm2015/lib/Key.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
export var Key = /*@__PURE__*/ (function (Key) {
    Key[Key["None"] = 0] = "None";
    Key[Key["Back"] = 1] = "Back";
    Key[Key["Tab"] = 2] = "Tab";
    Key[Key["Enter"] = 3] = "Enter";
    Key[Key["Shift"] = 4] = "Shift";
    Key[Key["Ctrl"] = 5] = "Ctrl";
    Key[Key["Alt"] = 6] = "Alt";
    Key[Key["CapsLock"] = 7] = "CapsLock";
    Key[Key["Escape"] = 8] = "Escape";
    Key[Key["Space"] = 9] = "Space";
    Key[Key["PageUp"] = 10] = "PageUp";
    Key[Key["PageDown"] = 11] = "PageDown";
    Key[Key["End"] = 12] = "End";
    Key[Key["Home"] = 13] = "Home";
    Key[Key["Left"] = 14] = "Left";
    Key[Key["Up"] = 15] = "Up";
    Key[Key["Right"] = 16] = "Right";
    Key[Key["Down"] = 17] = "Down";
    Key[Key["Insert"] = 18] = "Insert";
    Key[Key["del"] = 19] = "del";
    Key[Key["D0"] = 20] = "D0";
    Key[Key["D1"] = 21] = "D1";
    Key[Key["D2"] = 22] = "D2";
    Key[Key["D3"] = 23] = "D3";
    Key[Key["D4"] = 24] = "D4";
    Key[Key["D5"] = 25] = "D5";
    Key[Key["D6"] = 26] = "D6";
    Key[Key["D7"] = 27] = "D7";
    Key[Key["D8"] = 28] = "D8";
    Key[Key["D9"] = 29] = "D9";
    Key[Key["A"] = 30] = "A";
    Key[Key["B"] = 31] = "B";
    Key[Key["C"] = 32] = "C";
    Key[Key["D"] = 33] = "D";
    Key[Key["E"] = 34] = "E";
    Key[Key["F"] = 35] = "F";
    Key[Key["G"] = 36] = "G";
    Key[Key["H"] = 37] = "H";
    Key[Key["I"] = 38] = "I";
    Key[Key["J"] = 39] = "J";
    Key[Key["K"] = 40] = "K";
    Key[Key["L"] = 41] = "L";
    Key[Key["M"] = 42] = "M";
    Key[Key["N"] = 43] = "N";
    Key[Key["O"] = 44] = "O";
    Key[Key["P"] = 45] = "P";
    Key[Key["Q"] = 46] = "Q";
    Key[Key["R"] = 47] = "R";
    Key[Key["S"] = 48] = "S";
    Key[Key["T"] = 49] = "T";
    Key[Key["U"] = 50] = "U";
    Key[Key["V"] = 51] = "V";
    Key[Key["W"] = 52] = "W";
    Key[Key["X"] = 53] = "X";
    Key[Key["Y"] = 54] = "Y";
    Key[Key["Z"] = 55] = "Z";
    Key[Key["F1"] = 56] = "F1";
    Key[Key["F2"] = 57] = "F2";
    Key[Key["F3"] = 58] = "F3";
    Key[Key["F4"] = 59] = "F4";
    Key[Key["F5"] = 60] = "F5";
    Key[Key["F6"] = 61] = "F6";
    Key[Key["F7"] = 62] = "F7";
    Key[Key["F8"] = 63] = "F8";
    Key[Key["F9"] = 64] = "F9";
    Key[Key["F10"] = 65] = "F10";
    Key[Key["F11"] = 66] = "F11";
    Key[Key["F12"] = 67] = "F12";
    Key[Key["NumPad0"] = 68] = "NumPad0";
    Key[Key["NumPad1"] = 69] = "NumPad1";
    Key[Key["NumPad2"] = 70] = "NumPad2";
    Key[Key["NumPad3"] = 71] = "NumPad3";
    Key[Key["NumPad4"] = 72] = "NumPad4";
    Key[Key["NumPad5"] = 73] = "NumPad5";
    Key[Key["NumPad6"] = 74] = "NumPad6";
    Key[Key["NumPad7"] = 75] = "NumPad7";
    Key[Key["NumPad8"] = 76] = "NumPad8";
    Key[Key["NumPad9"] = 77] = "NumPad9";
    Key[Key["Multiply"] = 78] = "Multiply";
    Key[Key["Add"] = 79] = "Add";
    Key[Key["Subtract"] = 80] = "Subtract";
    Key[Key["Decimal"] = 81] = "Decimal";
    Key[Key["Divide"] = 82] = "Divide";
    Key[Key["OemSemicolon"] = 83] = "OemSemicolon";
    Key[Key["OemQuestion"] = 84] = "OemQuestion";
    Key[Key["OemPipe"] = 85] = "OemPipe";
    Key[Key["OemTilde"] = 86] = "OemTilde";
    Key[Key["OemPlus"] = 87] = "OemPlus";
    Key[Key["OemMinus"] = 88] = "OemMinus";
    Key[Key["Unknown"] = 255] = "Unknown";
    return Key;
})({});
/**
 * @hidden
 */
export let Key_$type = /*@__PURE__*/ markEnum('Key', 'None,0|Back,1|Tab,2|Enter,3|Shift,4|Ctrl,5|Alt,6|CapsLock,7|Escape,8|Space,9|PageUp,10|PageDown,11|End,12|Home,13|Left,14|Up,15|Right,16|Down,17|Insert,18|Delete:del,19|D0,20|D1,21|D2,22|D3,23|D4,24|D5,25|D6,26|D7,27|D8,28|D9,29|A,30|B,31|C,32|D,33|E,34|F,35|G,36|H,37|I,38|J,39|K,40|L,41|M,42|N,43|O,44|P,45|Q,46|R,47|S,48|T,49|U,50|V,51|W,52|X,53|Y,54|Z,55|F1,56|F2,57|F3,58|F4,59|F5,60|F6,61|F7,62|F8,63|F9,64|F10,65|F11,66|F12,67|NumPad0,68|NumPad1,69|NumPad2,70|NumPad3,71|NumPad4,72|NumPad5,73|NumPad6,74|NumPad7,75|NumPad8,76|NumPad9,77|Multiply,78|Add,79|Subtract,80|Decimal,81|Divide,82|OemSemicolon,83|OemQuestion,84|OemPipe,85|OemTilde,86|OemPlus,87|OemMinus,88|Unknown,255');
