// igniteui-react-core/esm2015/lib/LabelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { InputGroupItemDescription } from "./InputGroupItemDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let LabelDescription = /*@__PURE__*/ (() => {
    class LabelDescription extends InputGroupItemDescription {
        get_type() {
            return "Label";
        }
        constructor() {
            super();
            this.ac = null;
            this.ad = null;
            this.u = null;
            this.y = null;
            this.v = null;
            this.w = null;
            this.x = null;
            this.aq = null;
            this.al = null;
            this.am = null;
            this.ah = null;
            this.q = 0;
            this.ai = null;
            this.aj = null;
            this.an = null;
            this.ao = null;
            this.ae = null;
            this.af = null;
            this.ag = null;
            this.z = null;
            this.aa = null;
            this.s = 0;
            this.ak = null;
            this.ab = null;
            this.ap = null;
            this.l = false;
            this.m = false;
            this.k = false;
        }
        get baseTheme() {
            return this.ac;
        }
        set baseTheme(a) {
            this.ac = a;
            this.g("BaseTheme");
        }
        get density() {
            return this.ad;
        }
        set density(a) {
            this.ad = a;
            this.g("Density");
        }
        get actualDensity() {
            return this.u;
        }
        set actualDensity(a) {
            this.u = a;
            this.g("ActualDensity");
        }
        get actualTextColor() {
            return this.y;
        }
        set actualTextColor(a) {
            this.y = a;
            this.g("ActualTextColor");
        }
        get actualHighlightTextColor() {
            return this.v;
        }
        set actualHighlightTextColor(a) {
            this.v = a;
            this.g("ActualHighlightTextColor");
        }
        get actualHoverHighlightTextColor() {
            return this.w;
        }
        set actualHoverHighlightTextColor(a) {
            this.w = a;
            this.g("ActualHoverHighlightTextColor");
        }
        get actualHoverTextColor() {
            return this.x;
        }
        set actualHoverTextColor(a) {
            this.x = a;
            this.g("ActualHoverTextColor");
        }
        get textColor() {
            return this.aq;
        }
        set textColor(a) {
            this.aq = a;
            this.g("TextColor");
        }
        get highlightTextColor() {
            return this.al;
        }
        set highlightTextColor(a) {
            this.al = a;
            this.g("HighlightTextColor");
        }
        get hoverHighlightTextColor() {
            return this.am;
        }
        set hoverHighlightTextColor(a) {
            this.am = a;
            this.g("HoverHighlightTextColor");
        }
        get fontFamily() {
            return this.ah;
        }
        set fontFamily(a) {
            this.ah = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.q;
        }
        set fontSize(a) {
            this.q = a;
            this.g("FontSize");
        }
        get fontStyle() {
            return this.ai;
        }
        set fontStyle(a) {
            this.ai = a;
            this.g("FontStyle");
        }
        get fontWeight() {
            return this.aj;
        }
        set fontWeight(a) {
            this.aj = a;
            this.g("FontWeight");
        }
        get hoverTextColor() {
            return this.an;
        }
        set hoverTextColor(a) {
            this.an = a;
            this.g("HoverTextColor");
        }
        get id() {
            return this.ao;
        }
        set id(a) {
            this.ao = a;
            this.g("Id");
        }
        get display() {
            return this.ae;
        }
        set display(a) {
            this.ae = a;
            this.g("Display");
        }
        get flexDirection() {
            return this.af;
        }
        set flexDirection(a) {
            this.af = a;
            this.g("FlexDirection");
        }
        get flexGrow() {
            return this.ag;
        }
        set flexGrow(a) {
            this.ag = a;
            this.g("FlexGrow");
        }
        get alignItems() {
            return this.z;
        }
        set alignItems(a) {
            this.z = a;
            this.g("AlignItems");
        }
        get alignSelf() {
            return this.aa;
        }
        set alignSelf(a) {
            this.aa = a;
            this.g("AlignSelf");
        }
        get tabIndex() {
            return this.s;
        }
        set tabIndex(a) {
            this.s = a;
            this.g("TabIndex");
        }
        get $for() {
            return this.ak;
        }
        set $for(a) {
            this.ak = a;
            this.g("For");
        }
        get ariaLabel() {
            return this.ab;
        }
        set ariaLabel(a) {
            this.ab = a;
            this.g("AriaLabel");
        }
        get text() {
            return this.ap;
        }
        set text(a) {
            this.ap = a;
            this.g("Text");
        }
        get isHover() {
            return this.l;
        }
        set isHover(a) {
            this.l = a;
            this.g("IsHover");
        }
        get value() {
            return this.m;
        }
        set value(a) {
            this.m = a;
            this.g("Value");
        }
        get disabled() {
            return this.k;
        }
        set disabled(a) {
            this.k = a;
            this.g("Disabled");
        }
    }
    LabelDescription.$t = /*@__PURE__*/ markType(LabelDescription, 'LabelDescription', InputGroupItemDescription.$);
    return LabelDescription;
})();
