// igniteui-react-core/esm2015/lib/LabelCollisionInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let LabelCollisionInfo = /*@__PURE__*/ (() => {
    class LabelCollisionInfo extends Base {
        constructor() {
            super(...arguments);
            this.i = null;
            this.s = 0;
            this.t = 0;
            this.r = 0;
            this.q = 0;
            this.e = null;
            this.p = 0;
            this.o = 0;
            this.m = 0;
            this.n = 0;
            this.w = 0;
            this.x = 0;
            this.k = 0;
            this.l = 0;
            this.u = 0;
            this.v = 0;
            this.b = 0;
            this.c = null;
            this.y = null;
            this.f = false;
            this.a = null;
            this.d = null;
            this.g = false;
            this.j = null;
            this.h = false;
        }
    }
    LabelCollisionInfo.$t = /*@__PURE__*/ markType(LabelCollisionInfo, 'LabelCollisionInfo');
    return LabelCollisionInfo;
})();
