// igniteui-react-core/esm2015/lib/ListPanelTemplateWidthRequestedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ListPanelTemplateWidthRequestedEventArgsDescription = /*@__PURE__*/ (() => {
    class ListPanelTemplateWidthRequestedEventArgsDescription extends Description {
        get_type() {
            return "ListPanelTemplateWidthRequestedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = 0;
            this.l = null;
            this.h = 0;
        }
        get dataRow() {
            return this.j;
        }
        set dataRow(a) {
            this.j = a;
            this.g("DataRow");
        }
        get dataItemRef() {
            return this.l;
        }
        set dataItemRef(a) {
            this.l = a;
            this.g("DataItemRef");
        }
        get width() {
            return this.h;
        }
        set width(a) {
            this.h = a;
            this.g("Width");
        }
    }
    ListPanelTemplateWidthRequestedEventArgsDescription.$t = /*@__PURE__*/ markType(ListPanelTemplateWidthRequestedEventArgsDescription, 'ListPanelTemplateWidthRequestedEventArgsDescription', Description.$);
    return ListPanelTemplateWidthRequestedEventArgsDescription;
})();
