// igniteui-react-core/esm2015/lib/ListPanelItemModelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ListPanelItemModelDescription = /*@__PURE__*/ (() => {
    class ListPanelItemModelDescription extends Description {
        get_type() {
            return "ListPanelItemModel";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.v = 0;
            this.q = 0;
            this.x = null;
            this.j = false;
            this.p = 0;
            this.r = 0;
            this.h = false;
            this.i = false;
            this.k = false;
        }
        get dataRow() {
            return this.v;
        }
        set dataRow(a) {
            this.v = a;
            this.g("DataRow");
        }
        get rowHeight() {
            return this.q;
        }
        set rowHeight(a) {
            this.q = a;
            this.g("RowHeight");
        }
        get rowObjectRef() {
            return this.x;
        }
        set rowObjectRef(a) {
            this.x = a;
            this.g("RowObjectRef");
        }
        get isModelDirty() {
            return this.j;
        }
        set isModelDirty(a) {
            this.j = a;
            this.g("IsModelDirty");
        }
        get left() {
            return this.p;
        }
        set left(a) {
            this.p = a;
            this.g("Left");
        }
        get top() {
            return this.r;
        }
        set top(a) {
            this.r = a;
            this.g("Top");
        }
        get isActivated() {
            return this.h;
        }
        set isActivated(a) {
            this.h = a;
            this.g("IsActivated");
        }
        get isActivationSupported() {
            return this.i;
        }
        set isActivationSupported(a) {
            this.i = a;
            this.g("IsActivationSupported");
        }
        get isSelected() {
            return this.k;
        }
        set isSelected(a) {
            this.k = a;
            this.g("IsSelected");
        }
    }
    ListPanelItemModelDescription.$t = /*@__PURE__*/ markType(ListPanelItemModelDescription, 'ListPanelItemModelDescription', Description.$);
    return ListPanelItemModelDescription;
})();
