// igniteui-react-core/esm2015/lib/ListPanelActiveRowChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ListPanelActiveRowChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class ListPanelActiveRowChangedEventArgsDescription extends Description {
        get_type() {
            return "ListPanelActiveRowChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = 0;
            this.i = 0;
        }
        get newActiveRow() {
            return this.h;
        }
        set newActiveRow(a) {
            this.h = a;
            this.g("NewActiveRow");
        }
        get oldActiveRow() {
            return this.i;
        }
        set oldActiveRow(a) {
            this.i = a;
            this.g("OldActiveRow");
        }
    }
    ListPanelActiveRowChangedEventArgsDescription.$t = /*@__PURE__*/ markType(ListPanelActiveRowChangedEventArgsDescription, 'ListPanelActiveRowChangedEventArgsDescription', Description.$);
    return ListPanelActiveRowChangedEventArgsDescription;
})();
