// igniteui-react-core/esm2015/lib/ListPanelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ListPanelDescription = /*@__PURE__*/ (() => {
    class ListPanelDescription extends Description {
        get_type() {
            return "ListPanel";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.bc = null;
            this.t = false;
            this.a7 = null;
            this.a2 = null;
            this.ap = null;
            this.ae = 0;
            this.ag = 0;
            this.af = 0;
            this.ad = 0;
            this.aq = null;
            this.ba = null;
            this.i = null;
            this.h = null;
            this.k = null;
            this.l = null;
            this.ay = null;
            this.r = null;
            this.j = null;
            this.ai = 0;
            this.z = 0;
            this.au = null;
            this.aa = 0;
            this.av = null;
            this.aw = null;
            this.a6 = null;
            this.a5 = null;
            this.at = null;
            this.a3 = null;
            this.u = false;
            this.bd = null;
            this.ah = 0;
            this.v = false;
            this.a8 = null;
            this.a9 = null;
            this.bb = null;
            this.ar = null;
            this.ax = null;
            this.a4 = null;
            this.a1 = null;
            this.a0 = null;
            this.az = null;
            this.as = null;
        }
        get selectionMode() {
            return this.bc;
        }
        set selectionMode(a) {
            this.bc = a;
            this.g("SelectionMode");
        }
        get isActiveRowStyleEnabled() {
            return this.t;
        }
        set isActiveRowStyleEnabled(a) {
            this.t = a;
            this.g("IsActiveRowStyleEnabled");
        }
        get selectedBackground() {
            return this.a7;
        }
        set selectedBackground(a) {
            this.a7 = a;
            this.g("SelectedBackground");
        }
        get normalBackground() {
            return this.a2;
        }
        set normalBackground(a) {
            this.a2 = a;
            this.g("NormalBackground");
        }
        get activationBorder() {
            return this.ap;
        }
        set activationBorder(a) {
            this.ap = a;
            this.g("ActivationBorder");
        }
        get activationBorderLeftWidth() {
            return this.ae;
        }
        set activationBorderLeftWidth(a) {
            this.ae = a;
            this.g("ActivationBorderLeftWidth");
        }
        get activationBorderTopWidth() {
            return this.ag;
        }
        set activationBorderTopWidth(a) {
            this.ag = a;
            this.g("ActivationBorderTopWidth");
        }
        get activationBorderRightWidth() {
            return this.af;
        }
        set activationBorderRightWidth(a) {
            this.af = a;
            this.g("ActivationBorderRightWidth");
        }
        get activationBorderBottomWidth() {
            return this.ad;
        }
        set activationBorderBottomWidth(a) {
            this.ad = a;
            this.g("ActivationBorderBottomWidth");
        }
        get activationMode() {
            return this.aq;
        }
        set activationMode(a) {
            this.aq = a;
            this.g("ActivationMode");
        }
        get selectionBehavior() {
            return this.ba;
        }
        set selectionBehavior(a) {
            this.ba = a;
            this.g("SelectionBehavior");
        }
        get primaryKey() {
            return this.i;
        }
        set primaryKey(a) {
            this.i = a;
            this.g("PrimaryKey");
        }
        get actualPrimaryKey() {
            return this.h;
        }
        set actualPrimaryKey(a) {
            this.h = a;
            this.g("ActualPrimaryKey");
        }
        get selectedItems() {
            return this.k;
        }
        set selectedItems(a) {
            this.k = a;
            this.g("SelectedItems");
        }
        get selectedKeys() {
            return this.l;
        }
        set selectedKeys(a) {
            this.l = a;
            this.g("SelectedKeys");
        }
        get dataSourceRef() {
            return this.ay;
        }
        set dataSourceRef(a) {
            this.ay = a;
            this.g("DataSourceRef");
        }
        get dataSource() {
            return this.r;
        }
        set dataSource(a) {
            this.r = a;
            this.g("DataSource");
        }
        get schemaIncludedProperties() {
            return this.j;
        }
        set schemaIncludedProperties(a) {
            this.j = a;
            this.g("SchemaIncludedProperties");
        }
        get rowHeight() {
            return this.ai;
        }
        set rowHeight(a) {
            this.ai = a;
            this.g("RowHeight");
        }
        get actualRowHeight() {
            return this.z;
        }
        set actualRowHeight(a) {
            this.z = a;
            this.g("ActualRowHeight");
        }
        get cellFontFamily() {
            return this.au;
        }
        set cellFontFamily(a) {
            this.au = a;
            this.g("CellFontFamily");
        }
        get cellFontSize() {
            return this.aa;
        }
        set cellFontSize(a) {
            this.aa = a;
            this.g("CellFontSize");
        }
        get cellFontStyle() {
            return this.av;
        }
        set cellFontStyle(a) {
            this.av = a;
            this.g("CellFontStyle");
        }
        get cellFontWeight() {
            return this.aw;
        }
        set cellFontWeight(a) {
            this.aw = a;
            this.g("CellFontWeight");
        }
        get scrollbarStyle() {
            return this.a6;
        }
        set scrollbarStyle(a) {
            this.a6 = a;
            this.g("ScrollbarStyle");
        }
        get scrollbarBackground() {
            return this.a5;
        }
        set scrollbarBackground(a) {
            this.a5 = a;
            this.g("ScrollbarBackground");
        }
        get backgroundColor() {
            return this.at;
        }
        set backgroundColor(a) {
            this.at = a;
            this.g("BackgroundColor");
        }
        get orientation() {
            return this.a3;
        }
        set orientation(a) {
            this.a3 = a;
            this.g("Orientation");
        }
        get isCustomRowHeightEnabled() {
            return this.u;
        }
        set isCustomRowHeightEnabled(a) {
            this.u = a;
            this.g("IsCustomRowHeightEnabled");
        }
        get textColor() {
            return this.bd;
        }
        set textColor(a) {
            this.bd = a;
            this.g("TextColor");
        }
        get activeRow() {
            return this.ah;
        }
        set activeRow(a) {
            this.ah = a;
            this.g("ActiveRow");
        }
        get notifyOnAllSelectionChanges() {
            return this.v;
        }
        set notifyOnAllSelectionChanges(a) {
            this.v = a;
            this.g("NotifyOnAllSelectionChanges");
        }
        get selectedItemsChangedRef() {
            return this.a8;
        }
        set selectedItemsChangedRef(a) {
            this.a8 = a;
            this.g("SelectedItemsChangedRef");
        }
        get selectedKeysChangedRef() {
            return this.a9;
        }
        set selectedKeysChangedRef(a) {
            this.a9 = a;
            this.g("SelectedKeysChangedRef");
        }
        get selectionChangedRef() {
            return this.bb;
        }
        set selectionChangedRef(a) {
            this.bb = a;
            this.g("SelectionChangedRef");
        }
        get activeRowChangedRef() {
            return this.ar;
        }
        set activeRowChangedRef(a) {
            this.ar = a;
            this.g("ActiveRowChangedRef");
        }
        get contentRefreshedRef() {
            return this.ax;
        }
        set contentRefreshedRef(a) {
            this.ax = a;
            this.g("ContentRefreshedRef");
        }
        get rowUpdatingRef() {
            return this.a4;
        }
        set rowUpdatingRef(a) {
            this.a4 = a;
            this.g("RowUpdatingRef");
        }
        get itemWidthRequestedRef() {
            return this.a1;
        }
        set itemWidthRequestedRef(a) {
            this.a1 = a;
            this.g("ItemWidthRequestedRef");
        }
        get itemHeightRequestedRef() {
            return this.a0;
        }
        set itemHeightRequestedRef(a) {
            this.a0 = a;
            this.g("ItemHeightRequestedRef");
        }
        get itemClickedRef() {
            return this.az;
        }
        set itemClickedRef(a) {
            this.az = a;
            this.g("ItemClickedRef");
        }
        get actualPrimaryKeyChangeRef() {
            return this.as;
        }
        set actualPrimaryKeyChangeRef(a) {
            this.as = a;
            this.g("ActualPrimaryKeyChangeRef");
        }
    }
    ListPanelDescription.$t = /*@__PURE__*/ markType(ListPanelDescription, 'ListPanelDescription', Description.$);
    return ListPanelDescription;
})();
