// igniteui-react-core/esm2015/lib/ListPanelItemEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ListPanelItemEventArgsDescription = /*@__PURE__*/ (() => {
    class ListPanelItemEventArgsDescription extends Description {
        get_type() {
            return "ListPanelItemEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = null;
            this.m = false;
            this.l = false;
        }
        get itemInfo() {
            return this.j;
        }
        set itemInfo(a) {
            this.j = a;
            this.g("ItemInfo");
        }
        get listPanel() {
            return this.h;
        }
        set listPanel(a) {
            this.h = a;
            this.g("ListPanel");
        }
        get isLeftButton() {
            return this.m;
        }
        set isLeftButton(a) {
            this.m = a;
            this.g("IsLeftButton");
        }
        get isDoubleClick() {
            return this.l;
        }
        set isDoubleClick(a) {
            this.l = a;
            this.g("IsDoubleClick");
        }
    }
    ListPanelItemEventArgsDescription.$t = /*@__PURE__*/ markType(ListPanelItemEventArgsDescription, 'ListPanelItemEventArgsDescription', Description.$);
    return ListPanelItemEventArgsDescription;
})();
