// igniteui-react-core/esm2015/lib/MarkerSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let MarkerSeriesDescription = /*@__PURE__*/ (() => {
    class MarkerSeriesDescription extends SeriesDescription {
        get_type() {
            return "MarkerSeries";
        }
        constructor() {
            super();
            this.fg = null;
            this.fa = null;
            this.e3 = false;
            this.fe = null;
            this.fc = null;
            this.ff = null;
            this.e5 = 0;
            this.e9 = null;
            this.fb = null;
            this.e7 = null;
            this.fd = null;
            this.e8 = null;
        }
        get markerType() {
            return this.fg;
        }
        set markerType(a) {
            this.fg = a;
            this.g("MarkerType");
        }
        get actualMarkerType() {
            return this.fa;
        }
        set actualMarkerType(a) {
            this.fa = a;
            this.g("ActualMarkerType");
        }
        get isCustomMarkerCircular() {
            return this.e3;
        }
        set isCustomMarkerCircular(a) {
            this.e3 = a;
            this.g("IsCustomMarkerCircular");
        }
        get markerOutlineMode() {
            return this.fe;
        }
        set markerOutlineMode(a) {
            this.fe = a;
            this.g("MarkerOutlineMode");
        }
        get markerFillMode() {
            return this.fc;
        }
        set markerFillMode(a) {
            this.fc = a;
            this.g("MarkerFillMode");
        }
        get markerTemplateRef() {
            return this.ff;
        }
        set markerTemplateRef(a) {
            this.ff = a;
            this.g("MarkerTemplateRef");
        }
        get markerThickness() {
            return this.e5;
        }
        set markerThickness(a) {
            this.e5 = a;
            this.g("MarkerThickness");
        }
        get actualMarkerTemplateRef() {
            return this.e9;
        }
        set actualMarkerTemplateRef(a) {
            this.e9 = a;
            this.g("ActualMarkerTemplateRef");
        }
        get markerBrush() {
            return this.fb;
        }
        set markerBrush(a) {
            this.fb = a;
            this.g("MarkerBrush");
        }
        get actualMarkerBrush() {
            return this.e7;
        }
        set actualMarkerBrush(a) {
            this.e7 = a;
            this.g("ActualMarkerBrush");
        }
        get markerOutline() {
            return this.fd;
        }
        set markerOutline(a) {
            this.fd = a;
            this.g("MarkerOutline");
        }
        get actualMarkerOutline() {
            return this.e8;
        }
        set actualMarkerOutline(a) {
            this.e8 = a;
            this.g("ActualMarkerOutline");
        }
    }
    MarkerSeriesDescription.$t = /*@__PURE__*/ markType(MarkerSeriesDescription, 'MarkerSeriesDescription', SeriesDescription.$);
    return MarkerSeriesDescription;
})();
