// igniteui-react-core/esm2015/lib/MovingAverageConvergenceDivergenceIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicatorDescription } from "./StrategyBasedIndicatorDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let MovingAverageConvergenceDivergenceIndicatorDescription = /*@__PURE__*/ (() => {
    class MovingAverageConvergenceDivergenceIndicatorDescription extends StrategyBasedIndicatorDescription {
        get_type() {
            return "MovingAverageConvergenceDivergenceIndicator";
        }
        constructor() {
            super();
            this.gm = 0;
            this.gl = 0;
            this.gn = 0;
        }
        get shortPeriod() {
            return this.gm;
        }
        set shortPeriod(a) {
            this.gm = a;
            this.g("ShortPeriod");
        }
        get longPeriod() {
            return this.gl;
        }
        set longPeriod(a) {
            this.gl = a;
            this.g("LongPeriod");
        }
        get signalPeriod() {
            return this.gn;
        }
        set signalPeriod(a) {
            this.gn = a;
            this.g("SignalPeriod");
        }
    }
    MovingAverageConvergenceDivergenceIndicatorDescription.$t = /*@__PURE__*/ markType(MovingAverageConvergenceDivergenceIndicatorDescription, 'MovingAverageConvergenceDivergenceIndicatorDescription', StrategyBasedIndicatorDescription.$);
    return MovingAverageConvergenceDivergenceIndicatorDescription;
})();
