// igniteui-react-core/esm2015/lib/MultiSliderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let MultiSliderDescription = /*@__PURE__*/ (() => {
    class MultiSliderDescription extends Description {
        get_type() {
            return "MultiSlider";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.t = 0;
            this.s = 0;
            this.x = 0;
            this.ae = 0;
            this.af = 0;
            this.ag = 0;
            this.u = 0;
            this.n = 0;
            this.j = null;
            this.aj = 0;
            this.w = 0;
            this.r = 0;
            this.ad = 0;
            this.ac = 0;
            this.ai = 0;
            this.ah = 0;
            this.o = 0;
            this.bc = null;
            this.bh = null;
            this.aa = 0;
            this.bl = null;
            this.ab = 0;
            this.z = 0;
            this.a7 = null;
            this.a8 = null;
            this.p = 0;
            this.bd = null;
            this.be = null;
            this.v = 0;
            this.bf = null;
            this.bm = null;
            this.a9 = null;
            this.bb = null;
            this.ba = null;
            this.q = 0;
            this.l = false;
            this.bi = null;
            this.y = 0;
            this.bj = null;
            this.bk = null;
            this.bg = null;
            this.bo = null;
            this.bn = null;
            this.bq = null;
            this.bp = null;
        }
        get thumbs() {
            return this.h;
        }
        set thumbs(a) {
            this.h = a;
            this.g("Thumbs");
        }
        get min() {
            return this.t;
        }
        set min(a) {
            this.t = a;
            this.g("Min");
        }
        get max() {
            return this.s;
        }
        set max(a) {
            this.s = a;
            this.g("Max");
        }
        get step() {
            return this.x;
        }
        set step(a) {
            this.x = a;
            this.g("Step");
        }
        get yMax() {
            return this.ae;
        }
        set yMax(a) {
            this.ae = a;
            this.g("YMax");
        }
        get yMin() {
            return this.af;
        }
        set yMin(a) {
            this.af = a;
            this.g("YMin");
        }
        get yStep() {
            return this.ag;
        }
        set yStep(a) {
            this.ag = a;
            this.g("YStep");
        }
        get pixelScalingRatio() {
            return this.u;
        }
        set pixelScalingRatio(a) {
            this.u = a;
            this.g("PixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.n;
        }
        set actualPixelScalingRatio(a) {
            this.n = a;
            this.g("ActualPixelScalingRatio");
        }
        get windowRect() {
            return this.j;
        }
        set windowRect(a) {
            this.j = a;
            this.g("WindowRect");
        }
        get yValue() {
            return this.aj;
        }
        set yValue(a) {
            this.aj = a;
            this.g("YValue");
        }
        get startInset() {
            return this.w;
        }
        set startInset(a) {
            this.w = a;
            this.g("StartInset");
        }
        get endInset() {
            return this.r;
        }
        set endInset(a) {
            this.r = a;
            this.g("EndInset");
        }
        get trackStartInset() {
            return this.ad;
        }
        set trackStartInset(a) {
            this.ad = a;
            this.g("TrackStartInset");
        }
        get trackEndInset() {
            return this.ac;
        }
        set trackEndInset(a) {
            this.ac = a;
            this.g("TrackEndInset");
        }
        get yTrackStartInset() {
            return this.ai;
        }
        set yTrackStartInset(a) {
            this.ai = a;
            this.g("YTrackStartInset");
        }
        get yTrackEndInset() {
            return this.ah;
        }
        set yTrackEndInset(a) {
            this.ah = a;
            this.g("YTrackEndInset");
        }
        get barExtent() {
            return this.o;
        }
        set barExtent(a) {
            this.o = a;
            this.g("BarExtent");
        }
        get orientation() {
            return this.bc;
        }
        set orientation(a) {
            this.bc = a;
            this.g("Orientation");
        }
        get thumbBrush() {
            return this.bh;
        }
        set thumbBrush(a) {
            this.bh = a;
            this.g("ThumbBrush");
        }
        get thumbStrokeThickness() {
            return this.aa;
        }
        set thumbStrokeThickness(a) {
            this.aa = a;
            this.g("ThumbStrokeThickness");
        }
        get thumbOutline() {
            return this.bl;
        }
        set thumbOutline(a) {
            this.bl = a;
            this.g("ThumbOutline");
        }
        get thumbWidth() {
            return this.ab;
        }
        set thumbWidth(a) {
            this.ab = a;
            this.g("ThumbWidth");
        }
        get thumbHeight() {
            return this.z;
        }
        set thumbHeight(a) {
            this.z = a;
            this.g("ThumbHeight");
        }
        get barBrush() {
            return this.a7;
        }
        set barBrush(a) {
            this.a7 = a;
            this.g("BarBrush");
        }
        get barOutline() {
            return this.a8;
        }
        set barOutline(a) {
            this.a8 = a;
            this.g("BarOutline");
        }
        get barStrokeThickness() {
            return this.p;
        }
        set barStrokeThickness(a) {
            this.p = a;
            this.g("BarStrokeThickness");
        }
        get rangeThumbBrush() {
            return this.bd;
        }
        set rangeThumbBrush(a) {
            this.bd = a;
            this.g("RangeThumbBrush");
        }
        get rangeThumbOutline() {
            return this.be;
        }
        set rangeThumbOutline(a) {
            this.be = a;
            this.g("RangeThumbOutline");
        }
        get rangeThumbStrokeThickness() {
            return this.v;
        }
        set rangeThumbStrokeThickness(a) {
            this.v = a;
            this.g("RangeThumbStrokeThickness");
        }
        get rangeThumbRidgesBrush() {
            return this.bf;
        }
        set rangeThumbRidgesBrush(a) {
            this.bf = a;
            this.g("RangeThumbRidgesBrush");
        }
        get thumbRidgesBrush() {
            return this.bm;
        }
        set thumbRidgesBrush(a) {
            this.bm = a;
            this.g("ThumbRidgesBrush");
        }
        get calloutBrush() {
            return this.a9;
        }
        set calloutBrush(a) {
            this.a9 = a;
            this.g("CalloutBrush");
        }
        get calloutTextColor() {
            return this.bb;
        }
        set calloutTextColor(a) {
            this.bb = a;
            this.g("CalloutTextColor");
        }
        get calloutOutline() {
            return this.ba;
        }
        set calloutOutline(a) {
            this.ba = a;
            this.g("CalloutOutline");
        }
        get calloutStrokeThickness() {
            return this.q;
        }
        set calloutStrokeThickness(a) {
            this.q = a;
            this.g("CalloutStrokeThickness");
        }
        get areThumbCalloutsEnabled() {
            return this.l;
        }
        set areThumbCalloutsEnabled(a) {
            this.l = a;
            this.g("AreThumbCalloutsEnabled");
        }
        get thumbCalloutFontFamily() {
            return this.bi;
        }
        set thumbCalloutFontFamily(a) {
            this.bi = a;
            this.g("ThumbCalloutFontFamily");
        }
        get thumbCalloutFontSize() {
            return this.y;
        }
        set thumbCalloutFontSize(a) {
            this.y = a;
            this.g("ThumbCalloutFontSize");
        }
        get thumbCalloutFontStyle() {
            return this.bj;
        }
        set thumbCalloutFontStyle(a) {
            this.bj = a;
            this.g("ThumbCalloutFontStyle");
        }
        get thumbCalloutFontWeight() {
            return this.bk;
        }
        set thumbCalloutFontWeight(a) {
            this.bk = a;
            this.g("ThumbCalloutFontWeight");
        }
        get resolvingToolTipValueRef() {
            return this.bg;
        }
        set resolvingToolTipValueRef(a) {
            this.bg = a;
            this.g("ResolvingToolTipValueRef");
        }
        get thumbValueChangingRef() {
            return this.bo;
        }
        set thumbValueChangingRef(a) {
            this.bo = a;
            this.g("ThumbValueChangingRef");
        }
        get thumbValueChangedRef() {
            return this.bn;
        }
        set thumbValueChangedRef(a) {
            this.bn = a;
            this.g("ThumbValueChangedRef");
        }
        get yValueChangingRef() {
            return this.bq;
        }
        set yValueChangingRef(a) {
            this.bq = a;
            this.g("YValueChangingRef");
        }
        get yValueChangedRef() {
            return this.bp;
        }
        set yValueChangedRef(a) {
            this.bp = a;
            this.g("YValueChangedRef");
        }
    }
    MultiSliderDescription.$t = /*@__PURE__*/ markType(MultiSliderDescription, 'MultiSliderDescription', Description.$);
    return MultiSliderDescription;
})();
