// igniteui-react-core/esm2015/lib/NativeUIPopup.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIContainer } from "./NativeUIContainer";
import { NativeUIPopupAnimationType_$type } from "./NativeUIPopupAnimationType";
import { NativeUIPopupPointerPosition_$type } from "./NativeUIPopupPointerPosition";
import { typeGetValue, enumGetBox, EnumUtil, markType } from "./type";
/**
 * @hidden
 */
export let NativeUIPopup = /*@__PURE__*/ (() => {
    class NativeUIPopup extends NativeUIContainer {
        constructor() {
            super(...arguments);
            this.bx = null;
            this.bv = null;
            this.bt = null;
        }
        get br() {
            return this.d;
        }
        get b6() {
            return this.ag(22);
        }
        get b5() {
            return this.ag(59);
        }
        get b2() {
            return this.ag(60);
        }
        get b0() {
            return this.ag(25);
        }
        set b0(a) {
            this.a4(25, a);
        }
        get b9() {
            return typeGetValue(this.ag(26));
        }
        set b9(a) {
            this.a4(26, a);
        }
        get b1() {
            return this.ag(39);
        }
        set b1(a) {
            this.a4(39, a);
        }
        get bs() {
            return EnumUtil.getEnumValue(NativeUIPopupAnimationType_$type, this.ag(35));
        }
        set bs(a) {
            this.a4(35, enumGetBox(NativeUIPopupAnimationType_$type, a));
        }
        get cf() {
            return this.ag(23);
        }
        set cf(a) {
            this.a4(23, a);
        }
        get b7() {
            return this.ag(17);
        }
        set b7(a) {
            this.a4(17, a);
        }
        get_u() {
            return this.ag(36);
        }
        set_u(a) {
            this.a4(36, a);
        }
        get b3() {
            return this.ag(37);
        }
        set b3(a) {
            this.a4(37, a);
        }
        get ca() {
            return typeGetValue(this.ag(40));
        }
        set ca(a) {
            this.a4(40, a);
        }
        get b4() {
            return this.ag(41);
        }
        set b4(a) {
            this.a4(41, a);
        }
        get bz() {
            return EnumUtil.getEnumValue(NativeUIPopupPointerPosition_$type, this.ag(43));
        }
        set bz(a) {
            this.a4(43, enumGetBox(NativeUIPopupPointerPosition_$type, a));
        }
        get cg() {
            return this.ag(42);
        }
        set cg(a) {
            this.a4(42, a);
        }
        get b8() {
            return this.ag(61);
        }
        set b8(a) {
            this.a4(61, a);
        }
        get by() {
            return this.bx;
        }
        set by(a) {
            let b = this.bx;
            this.bx = a;
            if (this.bx != b) {
                if (b != null) {
                    this.a2(5, b);
                }
                if (this.bx != null) {
                    this.aq(5, this.bx);
                }
            }
        }
        get bw() {
            return this.bv;
        }
        set bw(a) {
            let b = this.bv;
            this.bv = a;
            if (this.bv != b) {
                if (b != null) {
                    this.a2(28, b);
                }
                if (this.bv != null) {
                    this.aq(28, this.bv);
                }
            }
        }
        get bu() {
            return this.bt;
        }
        set bu(a) {
            let b = this.bt;
            this.bt = a;
            if (this.bt != b) {
                if (b != null) {
                    this.a2(6, b);
                }
                if (this.bt != null) {
                    this.aq(6, this.bt);
                }
            }
        }
        ce(a, b, c) {
            this.br.showRelativeToExclusionRect(this.am, a, b, c);
        }
        cb() {
            this.br.close(this.am);
        }
        cc() {
            this.br.notifyContentSizeChanged(this.am);
        }
        cd(a) {
            this.br.notifyExclusionRectChanged(this.am, a);
        }
    }
    NativeUIPopup.$t = /*@__PURE__*/ markType(NativeUIPopup, 'NativeUIPopup', NativeUIContainer.$);
    return NativeUIPopup;
})();
