// igniteui-react-core/esm2015/lib/NativeUIDropDown.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { enumGetBox, EnumUtil, markType } from "./type";
/**
 * @hidden
 */
export let NativeUIDropDown = /*@__PURE__*/ (() => {
    class NativeUIDropDown extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bl = null;
        }
        get_e() {
            return 4;
        }
        get bk() {
            let a = this.ag(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        }
        set bk(a) {
            this.a4(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        }
        get bm() {
            return this.bl;
        }
        set bm(a) {
            let b = this.bl;
            this.bl = a;
            if (this.bl != b) {
                if (b != null) {
                    this.a2(3, b);
                }
                if (this.bl != null) {
                    this.aq(3, this.bl);
                }
            }
        }
        get bo() {
            let a = this.ag(9);
            return a;
        }
        set bo(a) {
            this.a4(9, a);
        }
        get bn() {
            let a = this.ag(8);
            return a;
        }
        set bn(a) {
            this.a4(8, a);
        }
    }
    NativeUIDropDown.$t = /*@__PURE__*/ markType(NativeUIDropDown, 'NativeUIDropDown', NativeUIComponent.$);
    return NativeUIDropDown;
})();
