// igniteui-react-core/esm2015/lib/NativeUI.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { UIElementPropertyAccessorFactory } from "./UIElementPropertyAccessorFactory";
import { MatrixTransform } from "./MatrixTransform";
import { TransformMatrixHelpler } from "./TransformMatrixHelpler";
/**
 * @hidden
 */
export let NativeUI = /*@__PURE__*/ (() => {
    class NativeUI extends Base {
        static ag(a, b, c) {
            let d = UIElementPropertyAccessorFactory.c.a(a, 0);
            d.c(b, c);
        }
        static l(a, b) {
            let c = UIElementPropertyAccessorFactory.c.a(a, 0);
            return c.b(b);
        }
        static y(a, b) {
            NativeUI.ag(1, a, b);
        }
        static g(a) {
            return NativeUI.l(1, a);
        }
        static m(a, b) {
            NativeUI.ag(2, a, b);
        }
        static aj(a) {
            return NativeUI.l(2, a);
        }
        static n(a, b) {
            NativeUI.ag(3, a, b);
        }
        static ak(a) {
            return NativeUI.l(3, a);
        }
        static p(a, b) {
            NativeUI.ag(4, a, b);
        }
        static ap(a) {
            return NativeUI.l(4, a);
        }
        static o(a, b) {
            NativeUI.ag(5, a, b);
        }
        static z(a, b) {
            NativeUI.ag(6, a, b);
        }
        static ar(a) {
            return NativeUI.l(6, a);
        }
        static ai(a) {
            return NativeUI.l(5, a);
        }
        static q(a, b) {
            NativeUI.ag(7, a, b);
        }
        static al(a) {
            return NativeUI.l(7, a);
        }
        static s(a, b) {
            NativeUI.ag(8, a, b);
        }
        static c(a) {
            return NativeUI.l(8, a);
        }
        static ah(a, b) {
            NativeUI.ag(9, a, b);
        }
        static k(a) {
            return NativeUI.l(9, a);
        }
        static x(a, b) {
            NativeUI.ag(26, a, b);
        }
        static f(a) {
            return NativeUI.l(26, a);
        }
        static t(a, b) {
            NativeUI.ag(10, a, b);
        }
        static d(a) {
            return NativeUI.l(10, a);
        }
        static v(a, b) {
            NativeUI.ag(11, a, b);
        }
        static ae(a, b) {
            NativeUI.ag(12, a, b);
        }
        static e(a) {
            return NativeUI.l(11, a);
        }
        static j(a) {
            return NativeUI.l(12, a);
        }
        static w(a, b) {
            NativeUI.ag(13, a, b);
        }
        static aq(a) {
            return NativeUI.l(13, a);
        }
        static af(a, b) {
            let c = TransformMatrixHelpler.f(b);
            let d = new MatrixTransform();
            d.j = c;
            NativeUI.ag(14, a, d);
        }
        static ao(a) {
            return NativeUI.l(14, a);
        }
        static aa(a, b) {
            NativeUI.ag(15, a, b);
        }
        static r(a, b) {
            NativeUI.ag(16, a, b);
        }
        static am(a) {
            return NativeUI.l(16, a);
        }
        static an(a) {
            return NativeUI.l(15, a);
        }
        static ad(a, b) {
            NativeUI.ag(17, a, b);
        }
        static i(a) {
            return NativeUI.l(17, a);
        }
        static ab(a, b) {
            NativeUI.ag(18, a, b);
        }
        static a(a) {
            return NativeUI.l(18, a);
        }
        static ac(a, b) {
            NativeUI.ag(19, a, b);
        }
        static h(a) {
            return NativeUI.l(19, a);
        }
        static u(a, b) {
            NativeUI.ag(21, a, b);
        }
        static b(a) {
            return NativeUI.l(21, a);
        }
    }
    NativeUI.$t = /*@__PURE__*/ markType(NativeUI, 'NativeUI');
    return NativeUI;
})();
