// igniteui-react-core/esm2015/lib/NativeUICheckbox.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { markType } from "./type";
/**
 * @hidden
 */
export let NativeUICheckbox = /*@__PURE__*/ (() => {
    class NativeUICheckbox extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bk = null;
        }
        get_e() {
            return 6;
        }
        get bm() {
            let a = this.ag(4);
            return a;
        }
        set bm(a) {
            this.a4(4, a);
        }
        get bn() {
            let a = this.ag(7);
            return a;
        }
        set bn(a) {
            this.a4(7, a);
        }
        get bo() {
            let a = this.ag(17);
            return a;
        }
        set bo(a) {
            this.a4(17, a);
        }
        get br() {
            return this.ag(65);
        }
        set br(a) {
            this.a4(65, a);
        }
        get bp() {
            return this.ag(66);
        }
        set bp(a) {
            this.a4(66, a);
        }
        get bs() {
            return this.ag(67);
        }
        set bs(a) {
            this.a4(67, a);
        }
        get bq() {
            return this.ag(68);
        }
        set bq(a) {
            this.a4(68, a);
        }
        get bl() {
            return this.bk;
        }
        set bl(a) {
            let b = this.bk;
            this.bk = a;
            if (this.bk != b) {
                if (b != null) {
                    this.a2(2, b);
                }
                if (this.bk != null) {
                    this.aq(2, this.bk);
                }
            }
        }
    }
    NativeUICheckbox.$t = /*@__PURE__*/ markType(NativeUICheckbox, 'NativeUICheckbox', NativeUIComponent.$);
    return NativeUICheckbox;
})();
