// igniteui-react-core/esm2015/lib/NativeUICombo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { enumGetBox, EnumUtil, markType } from "./type";
/**
 * @hidden
 */
export let NativeUICombo = /*@__PURE__*/ (() => {
    class NativeUICombo extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bl = null;
        }
        get_e() {
            return 19;
        }
        get bn() {
            let a = this.ag(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        }
        set bn(a) {
            this.a4(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        }
        get bm() {
            return this.bl;
        }
        set bm(a) {
            let b = this.bl;
            this.bl = a;
            if (this.bl != b) {
                if (b != null) {
                    this.a2(20, b);
                }
                if (this.bl != null) {
                    this.aq(20, this.bl);
                }
            }
        }
        get bk() {
            let a = this.ag(86);
            return a;
        }
        set bk(a) {
            this.a4(86, a);
        }
        get bp() {
            let a = this.ag(88);
            return a;
        }
        set bp(a) {
            this.a4(88, a);
        }
        get bq() {
            let a = this.ag(87);
            return a;
        }
        set bq(a) {
            this.a4(87, a);
        }
        get bo() {
            let a = this.ag(8);
            return a;
        }
        set bo(a) {
            this.a4(8, a);
        }
    }
    NativeUICombo.$t = /*@__PURE__*/ markType(NativeUICombo, 'NativeUICombo', NativeUIComponent.$);
    return NativeUICombo;
})();
