// igniteui-react-core/esm2015/lib/NativeUIInput.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIInputDisplayType_$type } from "./NativeUIInputDisplayType";
import { enumGetBox, EnumUtil, typeGetValue, markType } from "./type";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { NativeUIInputType_$type } from "./NativeUIInputType";
import { NativeUIComponentHorizontalAlignment_$type } from "./NativeUIComponentHorizontalAlignment";
/**
 * @hidden
 */
export let NativeUIInput = /*@__PURE__*/ (() => {
    class NativeUIInput extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bq = null;
            this.bw = null;
            this.by = null;
            this.br = null;
            this.bm = null;
            this.bk = null;
        }
        get bl() {
            return this.d;
        }
        get bu() {
            let a = this.ag(0);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIInputDisplayType_$type, a);
        }
        set bu(a) {
            this.a4(0, enumGetBox(NativeUIInputDisplayType_$type, a));
        }
        get bp() {
            let a = this.ag(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        }
        set bp(a) {
            this.a4(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        }
        get b5() {
            let a = this.ag(3);
            return a;
        }
        set b5(a) {
            this.a4(3, a);
        }
        get cb() {
            let a = this.ag(100);
            return a;
        }
        set cb(a) {
            this.a4(100, a);
        }
        get b7() {
            let a = this.ag(2);
            return a;
        }
        set b7(a) {
            this.a4(2, a);
        }
        get bs() {
            return this.bq;
        }
        set bs(a) {
            let b = this.bq;
            this.bq = a;
            if (this.bq != b) {
                if (b != null) {
                    this.a2(0, b);
                }
                if (this.bq != null) {
                    this.aq(0, this.bq);
                }
            }
        }
        get bx() {
            return this.bw;
        }
        set bx(a) {
            let b = this.bw;
            this.bw = a;
            if (this.bw != b) {
                if (b != null) {
                    this.a2(27, b);
                }
                if (this.bw != null) {
                    this.aq(27, this.bw);
                }
            }
        }
        get bz() {
            return this.by;
        }
        set bz(a) {
            let b = this.by;
            this.by = a;
            if (this.by != b) {
                if (b != null) {
                    this.a2(9, b);
                }
                if (this.by != null) {
                    this.aq(9, this.by);
                }
            }
        }
        get bt() {
            return this.br;
        }
        set bt(a) {
            let b = this.br;
            this.br = a;
            if (this.br != b) {
                if (b != null) {
                    this.a2(26, b);
                }
                if (this.bq != null) {
                    this.aq(26, this.bq);
                }
            }
        }
        get bn() {
            return this.bm;
        }
        set bn(a) {
            let b = this.bm;
            this.bm = a;
            if (this.bm != b) {
                if (b != null) {
                    this.a2(24, b);
                }
                if (this.bm != null) {
                    this.aq(24, this.bm);
                }
            }
        }
        get bv() {
            let a = this.ag(19);
            return a == null ? 0 : EnumUtil.getEnumValue(NativeUIInputType_$type, a);
        }
        set bv(a) {
            this.a4(19, enumGetBox(NativeUIInputType_$type, a));
        }
        get bo() {
            let a = this.ag(20);
            return a == null ? 0 : EnumUtil.getEnumValue(NativeUIComponentHorizontalAlignment_$type, a);
        }
        set bo(a) {
            this.a4(20, enumGetBox(NativeUIComponentHorizontalAlignment_$type, a));
        }
        get ce() {
            let a = this.ag(33);
            return a == null ? null : a;
        }
        set ce(a) {
            this.a4(33, a);
        }
        get b2() {
            let a = this.ag(21);
            return a == null ? false : a;
        }
        set b2(a) {
            this.a4(21, a);
        }
        get b0() {
            let a = this.ag(32);
            return a == null ? false : a;
        }
        set b0(a) {
            this.a4(32, a);
        }
        get b6() {
            let a = this.ag(97);
            return a;
        }
        set b6(a) {
            this.a4(97, a);
        }
        get cc() {
            let a = this.ag(98);
            return a;
        }
        set cc(a) {
            this.a4(98, a);
        }
        get cd() {
            let a = this.ag(99);
            return a;
        }
        set cd(a) {
            this.a4(99, a);
        }
        get b4() {
            let a = this.ag(102);
            return typeGetValue(a);
        }
        set b4(a) {
            this.a4(102, a);
        }
        get b3() {
            let a = this.ag(103);
            return typeGetValue(a);
        }
        set b3(a) {
            this.a4(103, a);
        }
        get b1() {
            let a = this.ag(104);
            return a;
        }
        set b1(a) {
            this.a4(104, a);
        }
        b8() {
            this.bl.resetIsExpanded(this.am);
        }
        b9() {
            this.bl.select(this.am);
        }
        ca(a, b) {
            this.bl.setSelectionRange(this.am, a, b);
        }
    }
    NativeUIInput.$t = /*@__PURE__*/ markType(NativeUIInput, 'NativeUIInput', NativeUIComponent.$);
    return NativeUIInput;
})();
