// igniteui-react-core/esm2015/lib/NativeUISlider.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { enumGetBox, EnumUtil, markType } from "./type";
import { NativeUISliderOrientation_$type } from "./NativeUISliderOrientation";
/**
 * @hidden
 */
export let NativeUISlider = /*@__PURE__*/ (() => {
    class NativeUISlider extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bl = null;
            this.bm = null;
        }
        get_e() {
            return 5;
        }
        get bs() {
            let a = this.ag(2);
            return a;
        }
        set bs(a) {
            this.a4(2, a);
        }
        get bq() {
            let a = this.ag(12);
            return a;
        }
        set bq(a) {
            this.a4(12, a);
        }
        get bu() {
            let a = this.ag(94);
            return a;
        }
        set bu(a) {
            this.a4(94, a);
        }
        get bv() {
            let a = this.ag(89);
            return a;
        }
        set bv(a) {
            this.a4(89, a);
        }
        get bx() {
            let a = this.ag(91);
            return a;
        }
        set bx(a) {
            this.a4(91, a);
        }
        get bz() {
            let a = this.ag(93);
            return a;
        }
        set bz(a) {
            this.a4(93, a);
        }
        get by() {
            let a = this.ag(92);
            return a;
        }
        set by(a) {
            this.a4(92, a);
        }
        get bw() {
            let a = this.ag(90);
            return a;
        }
        set bw(a) {
            this.a4(90, a);
        }
        get bk() {
            let a = this.ag(30);
            return EnumUtil.getEnumValue(NativeUISliderOrientation_$type, a);
        }
        set bk(a) {
            this.a4(30, enumGetBox(NativeUISliderOrientation_$type, a));
        }
        get bp() {
            let a = this.ag(13);
            return a;
        }
        set bp(a) {
            this.a4(13, a);
        }
        get bt() {
            let a = this.ag(95);
            return a;
        }
        set bt(a) {
            this.a4(95, a);
        }
        get br() {
            let a = this.ag(14);
            return a;
        }
        set br(a) {
            this.a4(14, a);
        }
        get bn() {
            return this.bl;
        }
        set bn(a) {
            let b = this.bl;
            this.bl = a;
            if (this.bl != b) {
                if (b != null) {
                    this.a2(4, b);
                }
                if (this.bl != null) {
                    this.aq(4, this.bl);
                }
            }
        }
        get bo() {
            return this.bm;
        }
        set bo(a) {
            let b = this.bm;
            this.bm = a;
            if (this.bm != b) {
                if (b != null) {
                    this.a2(0, b);
                }
                if (this.bm != null) {
                    this.aq(0, this.bm);
                }
            }
        }
    }
    NativeUISlider.$t = /*@__PURE__*/ markType(NativeUISlider, 'NativeUISlider', NativeUIComponent.$);
    return NativeUISlider;
})();
