// igniteui-react-core/esm2015/lib/NativeUIKeyEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let NativeUIKeyEventArgs = /*@__PURE__*/ (() => {
    class NativeUIKeyEventArgs extends Base {
        constructor() {
            super(...arguments);
            this.g = 0;
            this.h = null;
            this.e = false;
            this.b = false;
            this.a = false;
            this.f = false;
            this.d = false;
            this.c = false;
        }
    }
    NativeUIKeyEventArgs.$t = /*@__PURE__*/ markType(NativeUIKeyEventArgs, 'NativeUIKeyEventArgs');
    return NativeUIKeyEventArgs;
})();
