// igniteui-react-core/esm2015/lib/NativeUIWindow.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let NativeUIWindow = /*@__PURE__*/ (() => {
    class NativeUIWindow extends Base {
        static get b() {
            return NativeUIWindow.a;
        }
        static set b(a) {
            NativeUIWindow.a = a;
        }
        static g(a, b) {
            NativeUIWindow.b.addEventListener(a, b);
        }
        static h(a, b, c) {
            NativeUIWindow.b.addEventListener1(a, b, c);
        }
        static i(a, b) {
            NativeUIWindow.b.removeEventListener(a, b);
        }
        static e() {
            return NativeUIWindow.b.getWidth();
        }
        static f(a) {
            return NativeUIWindow.b.getWidth1(a);
        }
        static c() {
            return NativeUIWindow.b.getHeight();
        }
        static d(a) {
            return NativeUIWindow.b.getHeight1(a);
        }
    }
    NativeUIWindow.$t = /*@__PURE__*/ markType(NativeUIWindow, 'NativeUIWindow');
    NativeUIWindow.a = null;
    return NativeUIWindow;
})();
