// igniteui-react-core/esm2015/lib/NativeUIButton.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { enumGetBox, EnumUtil, markType } from "./type";
import { NativeUIButtonDisplayType_$type } from "./NativeUIButtonDisplayType";
/**
 * @hidden
 */
export let NativeUIButton = /*@__PURE__*/ (() => {
    class NativeUIButton extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bl = null;
        }
        get_e() {
            return 1;
        }
        get bp() {
            let a = this.ag(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        }
        set bp(a) {
            this.a4(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        }
        get bn() {
            let a = this.ag(0);
            return a == null ? 0 : EnumUtil.getEnumValue(NativeUIButtonDisplayType_$type, a);
        }
        set bn(a) {
            this.a4(0, enumGetBox(NativeUIButtonDisplayType_$type, a));
        }
        get by() {
            let a = this.ag(3);
            return a;
        }
        set by(a) {
            this.a4(3, a);
        }
        get bo() {
            return this.ag(54);
        }
        set bo(a) {
            this.a4(54, a);
        }
        get bu() {
            return this.ag(4);
        }
        set bu(a) {
            this.a4(4, a);
        }
        get bs() {
            return this.ag(5);
        }
        set bs(a) {
            this.a4(5, a);
        }
        get br() {
            return this.ag(6);
        }
        set br(a) {
            this.a4(6, a);
        }
        get bw() {
            return this.ag(32);
        }
        set bw(a) {
            this.a4(32, a);
        }
        get b0() {
            return this.ag(24);
        }
        set b0(a) {
            this.a4(24, a);
        }
        get b1() {
            return this.ag(55);
        }
        set b1(a) {
            this.a4(55, a);
        }
        get bv() {
            return this.ag(57);
        }
        set bv(a) {
            this.a4(57, a);
        }
        get bt() {
            return this.ag(56);
        }
        set bt(a) {
            this.a4(56, a);
        }
        get bq() {
            return this.ag(58);
        }
        set bq(a) {
            this.a4(58, a);
        }
        get bz() {
            return this.ag(17);
        }
        set bz(a) {
            this.a4(17, a);
        }
        get bk() {
            return this.ag(11);
        }
        set bk(a) {
            this.a4(11, a);
        }
        get bx() {
            return this.ag(85);
        }
        set bx(a) {
            this.a4(85, a);
        }
        get bm() {
            return this.bl;
        }
        set bm(a) {
            let b = this.bl;
            this.bl = a;
            if (this.bl != b) {
                if (b != null) {
                    this.a2(1, b);
                }
                if (this.bl != null) {
                    this.aq(1, this.bl);
                }
            }
        }
    }
    NativeUIButton.$t = /*@__PURE__*/ markType(NativeUIButton, 'NativeUIButton', NativeUIComponent.$);
    return NativeUIButton;
})();
