// igniteui-react-core/esm2015/lib/NativeUIColorEditor.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { markType } from "./type";
/**
 * @hidden
 */
export let NativeUIColorEditor = /*@__PURE__*/ (() => {
    class NativeUIColorEditor extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bk = null;
            this.bl = null;
        }
        get_e() {
            return 20;
        }
        get bo() {
            let a = this.ag(2);
            return a;
        }
        set bo(a) {
            this.a4(2, a);
        }
        get bm() {
            return this.bk;
        }
        set bm(a) {
            let b = this.bk;
            this.bk = a;
            if (this.bk != b) {
                if (b != null) {
                    this.a2(4, b);
                }
                if (this.bk != null) {
                    this.aq(4, this.bk);
                }
            }
        }
        get bn() {
            return this.bl;
        }
        set bn(a) {
            let b = this.bl;
            this.bl = a;
            if (this.bl != b) {
                if (b != null) {
                    this.a2(0, b);
                }
                if (this.bl != null) {
                    this.aq(0, this.bl);
                }
            }
        }
    }
    NativeUIColorEditor.$t = /*@__PURE__*/ markType(NativeUIColorEditor, 'NativeUIColorEditor', NativeUIComponent.$);
    return NativeUIColorEditor;
})();
