// igniteui-react-core/esm2015/lib/NumberFormatSpecifierDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FormatSpecifierDescription } from "./FormatSpecifierDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let NumberFormatSpecifierDescription = /*@__PURE__*/ (() => {
    class NumberFormatSpecifierDescription extends FormatSpecifierDescription {
        get_type() {
            return "NumberFormatSpecifier";
        }
        constructor() {
            super();
            this.ad = null;
            this.y = null;
            this.z = null;
            this.ab = null;
            this.ac = null;
            this.aa = null;
            this.ae = null;
            this.af = null;
            this.ag = null;
            this.ah = null;
            this.ai = null;
            this.aj = null;
            this.ak = null;
            this.l = false;
            this.q = 0;
            this.p = 0;
            this.n = 0;
            this.r = 0;
            this.o = 0;
        }
        get locale() {
            return this.ad;
        }
        set locale(a) {
            this.ad = a;
            this.g("Locale");
        }
        get compactDisplay() {
            return this.y;
        }
        set compactDisplay(a) {
            this.y = a;
            this.g("CompactDisplay");
        }
        get currency() {
            return this.z;
        }
        set currency(a) {
            this.z = a;
            this.g("Currency");
        }
        get currencyDisplay() {
            return this.ab;
        }
        set currencyDisplay(a) {
            this.ab = a;
            this.g("CurrencyDisplay");
        }
        get currencySign() {
            return this.ac;
        }
        set currencySign(a) {
            this.ac = a;
            this.g("CurrencySign");
        }
        get currencyCode() {
            return this.aa;
        }
        set currencyCode(a) {
            this.aa = a;
            this.g("CurrencyCode");
        }
        get localeMatcher() {
            return this.ae;
        }
        set localeMatcher(a) {
            this.ae = a;
            this.g("LocaleMatcher");
        }
        get notation() {
            return this.af;
        }
        set notation(a) {
            this.af = a;
            this.g("Notation");
        }
        get numberingSystem() {
            return this.ag;
        }
        set numberingSystem(a) {
            this.ag = a;
            this.g("NumberingSystem");
        }
        get signDisplay() {
            return this.ah;
        }
        set signDisplay(a) {
            this.ah = a;
            this.g("SignDisplay");
        }
        get style() {
            return this.ai;
        }
        set style(a) {
            this.ai = a;
            this.g("Style");
        }
        get unit() {
            return this.aj;
        }
        set unit(a) {
            this.aj = a;
            this.g("Unit");
        }
        get unitDisplay() {
            return this.ak;
        }
        set unitDisplay(a) {
            this.ak = a;
            this.g("UnitDisplay");
        }
        get useGrouping() {
            return this.l;
        }
        set useGrouping(a) {
            this.l = a;
            this.g("UseGrouping");
        }
        get minimumIntegerDigits() {
            return this.q;
        }
        set minimumIntegerDigits(a) {
            this.q = a;
            this.g("MinimumIntegerDigits");
        }
        get minimumFractionDigits() {
            return this.p;
        }
        set minimumFractionDigits(a) {
            this.p = a;
            this.g("MinimumFractionDigits");
        }
        get maximumFractionDigits() {
            return this.n;
        }
        set maximumFractionDigits(a) {
            this.n = a;
            this.g("MaximumFractionDigits");
        }
        get minimumSignificantDigits() {
            return this.r;
        }
        set minimumSignificantDigits(a) {
            this.r = a;
            this.g("MinimumSignificantDigits");
        }
        get maximumSignificantDigits() {
            return this.o;
        }
        set maximumSignificantDigits(a) {
            this.o = a;
            this.g("MaximumSignificantDigits");
        }
    }
    NumberFormatSpecifierDescription.$t = /*@__PURE__*/ markType(NumberFormatSpecifierDescription, 'NumberFormatSpecifierDescription', FormatSpecifierDescription.$);
    NumberFormatSpecifierDescription.__marshalByValue1 = true;
    NumberFormatSpecifierDescription.__marshalByValueAlias1 = "NumberFormatSpecifier";
    return NumberFormatSpecifierDescription;
})();
