// igniteui-react-core/esm2015/lib/NumericCellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CellInfoDescription } from "./CellInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let NumericCellInfoDescription = /*@__PURE__*/ (() => {
    class NumericCellInfoDescription extends CellInfoDescription {
        get_type() {
            return "NumericCellInfo";
        }
        constructor() {
            super();
            this.fs = 0;
            this.fo = false;
            this.f2 = null;
            this.fl = null;
            this.f1 = null;
            this.f3 = null;
            this.f5 = null;
            this.f4 = null;
            this.f6 = null;
            this.fu = 0;
            this.fv = 0;
            this.fw = 0;
            this.fp = false;
        }
        get numericValue() {
            return this.fs;
        }
        set numericValue(a) {
            this.fs = a;
            this.g("NumericValue");
        }
        get hasDecimalValue() {
            return this.fo;
        }
        set hasDecimalValue(a) {
            this.fo = a;
            this.g("HasDecimalValue");
        }
        get formatStringOverride() {
            return this.f2;
        }
        set formatStringOverride(a) {
            this.f2 = a;
            this.g("FormatStringOverride");
        }
        get formatSpecifiers() {
            return this.fl;
        }
        set formatSpecifiers(a) {
            this.fl = a;
            this.g("FormatSpecifiers");
        }
        get formatOverrideRef() {
            return this.f1;
        }
        set formatOverrideRef(a) {
            this.f1 = a;
            this.g("FormatOverrideRef");
        }
        get negativePrefix() {
            return this.f3;
        }
        set negativePrefix(a) {
            this.f3 = a;
            this.g("NegativePrefix");
        }
        get positivePrefix() {
            return this.f5;
        }
        set positivePrefix(a) {
            this.f5 = a;
            this.g("PositivePrefix");
        }
        get negativeSuffix() {
            return this.f4;
        }
        set negativeSuffix(a) {
            this.f4 = a;
            this.g("NegativeSuffix");
        }
        get positiveSuffix() {
            return this.f6;
        }
        set positiveSuffix(a) {
            this.f6 = a;
            this.g("PositiveSuffix");
        }
        get maxFractionDigits() {
            return this.fu;
        }
        set maxFractionDigits(a) {
            this.fu = a;
            this.g("MaxFractionDigits");
        }
        get minFractionDigits() {
            return this.fv;
        }
        set minFractionDigits(a) {
            this.fv = a;
            this.g("MinFractionDigits");
        }
        get minIntegerDigits() {
            return this.fw;
        }
        set minIntegerDigits(a) {
            this.fw = a;
            this.g("MinIntegerDigits");
        }
        get showGroupingSeparator() {
            return this.fp;
        }
        set showGroupingSeparator(a) {
            this.fp = a;
            this.g("ShowGroupingSeparator");
        }
    }
    NumericCellInfoDescription.$t = /*@__PURE__*/ markType(NumericCellInfoDescription, 'NumericCellInfoDescription', CellInfoDescription.$);
    NumericCellInfoDescription.__marshalByValue1 = true;
    NumericCellInfoDescription.__marshalByValueAlias1 = "NumericCellModel";
    return NumericCellInfoDescription;
})();
