// igniteui-react-core/esm2015/lib/NumericRadiusAxisDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NumericAxisBaseDescription } from "./NumericAxisBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let NumericRadiusAxisDescription = /*@__PURE__*/ (() => {
    class NumericRadiusAxisDescription extends NumericAxisBaseDescription {
        get_type() {
            return "NumericRadiusAxis";
        }
        constructor() {
            super();
            this.e8 = 0;
            this.e7 = 0;
        }
        get radiusExtentScale() {
            return this.e8;
        }
        set radiusExtentScale(a) {
            this.e8 = a;
            this.g("RadiusExtentScale");
        }
        get innerRadiusExtentScale() {
            return this.e7;
        }
        set innerRadiusExtentScale(a) {
            this.e7 = a;
            this.g("InnerRadiusExtentScale");
        }
    }
    NumericRadiusAxisDescription.$t = /*@__PURE__*/ markType(NumericRadiusAxisDescription, 'NumericRadiusAxisDescription', NumericAxisBaseDescription.$);
    return NumericRadiusAxisDescription;
})();
