// igniteui-react-core/esm2015/lib/ODataDataSourceSchema.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { IDataSourceSchema_$type } from "./IDataSourceSchema";
/**
 * @hidden
 */
export let ODataDataSourceSchema = /*@__PURE__*/ (() => {
    class ODataDataSourceSchema extends Base {
        constructor(a, b, c) {
            super();
            this.a = null;
            this.d = null;
            this.e = null;
            this.b = null;
            this.c = null;
            this.d = a;
            this.e = b;
            this.a = c;
        }
        get propertyNames() {
            return this.d;
        }
        get primaryKey() {
            return this.a;
        }
        get propertyTypes() {
            return this.e;
        }
        get propertyDataIntents() {
            return this.b;
        }
        get subSchemas() {
            return this.c;
        }
    }
    ODataDataSourceSchema.$t = /*@__PURE__*/ markType(ODataDataSourceSchema, 'ODataDataSourceSchema', Base.$, [IDataSourceSchema_$type]);
    return ODataDataSourceSchema;
})();
