// igniteui-react-core/esm2015/lib/OperatorSelectorValueChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let OperatorSelectorValueChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class OperatorSelectorValueChangedEventArgsDescription extends Description {
        get_type() {
            return "OperatorSelectorValueChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
        }
        get oldValue() {
            return this.i;
        }
        set oldValue(a) {
            this.i = a;
            this.g("OldValue");
        }
        get newValue() {
            return this.h;
        }
        set newValue(a) {
            this.h = a;
            this.g("NewValue");
        }
    }
    OperatorSelectorValueChangedEventArgsDescription.$t = /*@__PURE__*/ markType(OperatorSelectorValueChangedEventArgsDescription, 'OperatorSelectorValueChangedEventArgsDescription', Description.$);
    return OperatorSelectorValueChangedEventArgsDescription;
})();
