// igniteui-react-core/esm2015/lib/OperationFilterExpressionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FilterExpressionDescription } from "./FilterExpressionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let OperationFilterExpressionDescription = /*@__PURE__*/ (() => {
    class OperationFilterExpressionDescription extends FilterExpressionDescription {
        get_type() {
            return "OperationFilterExpression";
        }
        constructor() {
            super();
            this.r = null;
            this.s = null;
            this.k = null;
            this.l = null;
            this.q = null;
        }
        get propertyName() {
            return this.r;
        }
        set propertyName(a) {
            this.r = a;
            this.g("PropertyName");
        }
        get valueRef() {
            return this.s;
        }
        set valueRef(a) {
            this.s = a;
            this.g("ValueRef");
        }
        get left() {
            return this.k;
        }
        set left(a) {
            this.k = a;
            this.g("Left");
        }
        get right() {
            return this.l;
        }
        set right(a) {
            this.l = a;
            this.g("Right");
        }
        get operator() {
            return this.q;
        }
        set operator(a) {
            this.q = a;
            this.g("Operator");
        }
    }
    OperationFilterExpressionDescription.$t = /*@__PURE__*/ markType(OperationFilterExpressionDescription, 'OperationFilterExpressionDescription', FilterExpressionDescription.$);
    OperationFilterExpressionDescription.__marshalByValue1 = true;
    OperationFilterExpressionDescription.__marshalByValueAlias1 = "OperationFilterExpression";
    return OperationFilterExpressionDescription;
})();
