// igniteui-react-core/esm2015/lib/PageChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let PageChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class PageChangedEventArgsDescription extends Description {
        get_type() {
            return "PageChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = 0;
            this.j = null;
        }
        get page() {
            return this.h;
        }
        set page(a) {
            this.h = a;
            this.g("Page");
        }
        get dataRef() {
            return this.j;
        }
        set dataRef(a) {
            this.j = a;
            this.g("DataRef");
        }
    }
    PageChangedEventArgsDescription.$t = /*@__PURE__*/ markType(PageChangedEventArgsDescription, 'PageChangedEventArgsDescription', Description.$);
    return PageChangedEventArgsDescription;
})();
