// igniteui-react-core/esm2015/lib/PanelBridge.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIElementBridge } from "./NativeUIElementBridge";
import { List$1 } from "./List$1";
import { NativeUIComponent } from "./NativeUIComponent";
import { markType } from "./type";
/**
 * @hidden
 */
export let PanelBridge = /*@__PURE__*/ (() => {
    class PanelBridge extends NativeUIElementBridge {
        constructor() {
            super();
            this.t = null;
            this.t = new List$1(NativeUIComponent.$, 0);
        }
        v(a, b) {
            this.t.add(b);
            a.append(b.an);
        }
        w(a, b, c) {
            this.t.insert(b, c);
            if (this.t.count - 1 >= (b + 1)) {
                this.t._inner[b + 1].an.before(c.an);
            }
            else {
                a.append(c.an);
            }
        }
        s(a, b) {
            return this.t._inner[b];
        }
        u(a) {
            return a.getChildCount();
        }
        x(a, b) {
            this.t.remove(b);
            a.removeChild(b.an);
        }
        y(a, b) {
            this.t.removeAt(b);
            a.getChildAt(b).remove();
        }
    }
    PanelBridge.$t = /*@__PURE__*/ markType(PanelBridge, 'PanelBridge', NativeUIElementBridge.$);
    return PanelBridge;
})();
