// igniteui-react-core/esm2015/lib/PercentageVolumeOscillatorIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicatorDescription } from "./StrategyBasedIndicatorDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let PercentageVolumeOscillatorIndicatorDescription = /*@__PURE__*/ (() => {
    class PercentageVolumeOscillatorIndicatorDescription extends StrategyBasedIndicatorDescription {
        get_type() {
            return "PercentageVolumeOscillatorIndicator";
        }
        constructor() {
            super();
            this.gm = 0;
            this.gl = 0;
        }
        get shortPeriod() {
            return this.gm;
        }
        set shortPeriod(a) {
            this.gm = a;
            this.g("ShortPeriod");
        }
        get longPeriod() {
            return this.gl;
        }
        set longPeriod(a) {
            this.gl = a;
            this.g("LongPeriod");
        }
    }
    PercentageVolumeOscillatorIndicatorDescription.$t = /*@__PURE__*/ markType(PercentageVolumeOscillatorIndicatorDescription, 'PercentageVolumeOscillatorIndicatorDescription', StrategyBasedIndicatorDescription.$);
    return PercentageVolumeOscillatorIndicatorDescription;
})();
