// igniteui-react-core/esm2015/lib/PieChartBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let PieChartBaseDescription = /*@__PURE__*/ (() => {
    class PieChartBaseDescription extends Description {
        get tooltipTemplateRef() {
            return this.cm;
        }
        set tooltipTemplateRef(a) {
            this.cm = a;
            this.g("TooltipTemplateRef");
        }
        get fontFamily() {
            return this.bk;
        }
        set fontFamily(a) {
            this.bk = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.ap;
        }
        set fontSize(a) {
            this.ap = a;
            this.g("FontSize");
        }
        get fontWeight() {
            return this.bm;
        }
        set fontWeight(a) {
            this.bm = a;
            this.g("FontWeight");
        }
        get fontStyle() {
            return this.bl;
        }
        set fontStyle(a) {
            this.bl = a;
            this.g("FontStyle");
        }
        get_type() {
            return "PieChartBase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.cm = null;
            this.bk = null;
            this.ap = 0;
            this.bm = null;
            this.bl = null;
            this.bj = null;
            this.aq = 0;
            this.cn = null;
            this.bs = null;
            this.b1 = null;
            this.bu = null;
            this.bt = null;
            this.br = null;
            this.bi = null;
            this.bh = null;
            this.bw = null;
            this.bv = null;
            this.as = 0;
            this.av = 0;
            this.b7 = null;
            this.b6 = null;
            this.ao = 0;
            this.ax = 0;
            this.ac = false;
            this.ab = false;
            this.k = null;
            this.b3 = null;
            this.ar = 0;
            this.a0 = 0;
            this.ck = null;
            this.b4 = null;
            this.b5 = null;
            this.au = 0;
            this.at = 0;
            this.cd = null;
            this.ce = null;
            this.az = 0;
            this.ay = 0;
            this.j = null;
            this.h = null;
            this.p = null;
            this.i = null;
            this.bq = null;
            this.l = null;
            this.b8 = null;
            this.o = null;
            this.b0 = null;
            this.m = null;
            this.b2 = null;
            this.n = null;
            this.bz = null;
            this.by = null;
            this.ad = false;
            this.ae = false;
            this.af = false;
            this.ag = false;
            this.bx = null;
            this.bn = null;
            this.bo = null;
            this.aw = 0;
            this.an = 0;
            this.cf = null;
            this.bf = null;
            this.q = null;
            this.cl = null;
            this.bp = null;
            this.ca = null;
            this.cc = null;
            this.b9 = null;
            this.cb = null;
            this.cg = null;
            this.ch = null;
            this.cj = null;
            this.ci = null;
        }
        get dataSourceRef() {
            return this.bj;
        }
        set dataSourceRef(a) {
            this.bj = a;
            this.g("DataSourceRef");
        }
        get innerExtent() {
            return this.aq;
        }
        set innerExtent(a) {
            this.aq = a;
            this.g("InnerExtent");
        }
        get valueMemberPath() {
            return this.cn;
        }
        set valueMemberPath(a) {
            this.cn = a;
            this.g("ValueMemberPath");
        }
        get labelMemberPath() {
            return this.bs;
        }
        set labelMemberPath(a) {
            this.bs = a;
            this.g("LabelMemberPath");
        }
        get legendLabelMemberPath() {
            return this.b1;
        }
        set legendLabelMemberPath(a) {
            this.b1 = a;
            this.g("LegendLabelMemberPath");
        }
        get labelsPosition() {
            return this.bu;
        }
        set labelsPosition(a) {
            this.bu = a;
            this.g("LabelsPosition");
        }
        get labelOuterColor() {
            return this.bt;
        }
        set labelOuterColor(a) {
            this.bt = a;
            this.g("LabelOuterColor");
        }
        get labelInnerColor() {
            return this.br;
        }
        set labelInnerColor(a) {
            this.br = a;
            this.g("LabelInnerColor");
        }
        get actualLabelOuterColor() {
            return this.bi;
        }
        set actualLabelOuterColor(a) {
            this.bi = a;
            this.g("ActualLabelOuterColor");
        }
        get actualLabelInnerColor() {
            return this.bh;
        }
        set actualLabelInnerColor(a) {
            this.bh = a;
            this.g("ActualLabelInnerColor");
        }
        get leaderLineVisibility() {
            return this.bw;
        }
        set leaderLineVisibility(a) {
            this.bw = a;
            this.g("LeaderLineVisibility");
        }
        get leaderLineType() {
            return this.bv;
        }
        set leaderLineType(a) {
            this.bv = a;
            this.g("LeaderLineType");
        }
        get leaderLineMargin() {
            return this.as;
        }
        set leaderLineMargin(a) {
            this.as = a;
            this.g("LeaderLineMargin");
        }
        get othersCategoryThreshold() {
            return this.av;
        }
        set othersCategoryThreshold(a) {
            this.av = a;
            this.g("OthersCategoryThreshold");
        }
        get othersCategoryType() {
            return this.b7;
        }
        set othersCategoryType(a) {
            this.b7 = a;
            this.g("OthersCategoryType");
        }
        get othersCategoryText() {
            return this.b6;
        }
        set othersCategoryText(a) {
            this.b6 = a;
            this.g("OthersCategoryText");
        }
        get explodedRadius() {
            return this.ao;
        }
        set explodedRadius(a) {
            this.ao = a;
            this.g("ExplodedRadius");
        }
        get radiusFactor() {
            return this.ax;
        }
        set radiusFactor(a) {
            this.ax = a;
            this.g("RadiusFactor");
        }
        get allowSliceSelection() {
            return this.ac;
        }
        set allowSliceSelection(a) {
            this.ac = a;
            this.g("AllowSliceSelection");
        }
        get allowSliceExplosion() {
            return this.ab;
        }
        set allowSliceExplosion(a) {
            this.ab = a;
            this.g("AllowSliceExplosion");
        }
        get explodedSlices() {
            return this.k;
        }
        set explodedSlices(a) {
            this.k = a;
            this.g("ExplodedSlices");
        }
        get legendRef() {
            return this.b3;
        }
        set legendRef(a) {
            this.b3 = a;
            this.g("LegendRef");
        }
        get labelExtent() {
            return this.ar;
        }
        set labelExtent(a) {
            this.ar = a;
            this.g("LabelExtent");
        }
        get startAngle() {
            return this.a0;
        }
        set startAngle(a) {
            this.a0 = a;
            this.g("StartAngle");
        }
        get sweepDirection() {
            return this.ck;
        }
        set sweepDirection(a) {
            this.ck = a;
            this.g("SweepDirection");
        }
        get othersCategoryFill() {
            return this.b4;
        }
        set othersCategoryFill(a) {
            this.b4 = a;
            this.g("OthersCategoryFill");
        }
        get othersCategoryStroke() {
            return this.b5;
        }
        set othersCategoryStroke(a) {
            this.b5 = a;
            this.g("OthersCategoryStroke");
        }
        get othersCategoryStrokeThickness() {
            return this.au;
        }
        set othersCategoryStrokeThickness(a) {
            this.au = a;
            this.g("OthersCategoryStrokeThickness");
        }
        get othersCategoryOpacity() {
            return this.at;
        }
        set othersCategoryOpacity(a) {
            this.at = a;
            this.g("OthersCategoryOpacity");
        }
        get selectedSliceFill() {
            return this.cd;
        }
        set selectedSliceFill(a) {
            this.cd = a;
            this.g("SelectedSliceFill");
        }
        get selectedSliceStroke() {
            return this.ce;
        }
        set selectedSliceStroke(a) {
            this.ce = a;
            this.g("SelectedSliceStroke");
        }
        get selectedSliceStrokeThickness() {
            return this.az;
        }
        set selectedSliceStrokeThickness(a) {
            this.az = a;
            this.g("SelectedSliceStrokeThickness");
        }
        get selectedSliceOpacity() {
            return this.ay;
        }
        set selectedSliceOpacity(a) {
            this.ay = a;
            this.g("SelectedSliceOpacity");
        }
        get brushes() {
            return this.j;
        }
        set brushes(a) {
            this.j = a;
            this.g("Brushes");
        }
        get actualBrushes() {
            return this.h;
        }
        set actualBrushes(a) {
            this.h = a;
            this.g("ActualBrushes");
        }
        get outlines() {
            return this.p;
        }
        set outlines(a) {
            this.p = a;
            this.g("Outlines");
        }
        get actualOutlines() {
            return this.i;
        }
        set actualOutlines(a) {
            this.i = a;
            this.g("ActualOutlines");
        }
        get labelFormat() {
            return this.bq;
        }
        set labelFormat(a) {
            this.bq = a;
            this.g("LabelFormat");
        }
        get labelFormatSpecifiers() {
            return this.l;
        }
        set labelFormatSpecifiers(a) {
            this.l = a;
            this.g("LabelFormatSpecifiers");
        }
        get othersLabelFormat() {
            return this.b8;
        }
        set othersLabelFormat(a) {
            this.b8 = a;
            this.g("OthersLabelFormat");
        }
        get othersLabelFormatSpecifiers() {
            return this.o;
        }
        set othersLabelFormatSpecifiers(a) {
            this.o = a;
            this.g("OthersLabelFormatSpecifiers");
        }
        get legendLabelFormat() {
            return this.b0;
        }
        set legendLabelFormat(a) {
            this.b0 = a;
            this.g("LegendLabelFormat");
        }
        get legendLabelFormatSpecifiers() {
            return this.m;
        }
        set legendLabelFormatSpecifiers(a) {
            this.m = a;
            this.g("LegendLabelFormatSpecifiers");
        }
        get legendOthersLabelFormat() {
            return this.b2;
        }
        set legendOthersLabelFormat(a) {
            this.b2 = a;
            this.g("LegendOthersLabelFormat");
        }
        get legendOthersLabelFormatSpecifiers() {
            return this.n;
        }
        set legendOthersLabelFormatSpecifiers(a) {
            this.n = a;
            this.g("LegendOthersLabelFormatSpecifiers");
        }
        get legendItemTemplateRef() {
            return this.bz;
        }
        set legendItemTemplateRef(a) {
            this.bz = a;
            this.g("LegendItemTemplateRef");
        }
        get legendItemBadgeTemplateRef() {
            return this.by;
        }
        set legendItemBadgeTemplateRef(a) {
            this.by = a;
            this.g("LegendItemBadgeTemplateRef");
        }
        get isDragInteractionEnabled() {
            return this.ad;
        }
        set isDragInteractionEnabled(a) {
            this.ad = a;
            this.g("IsDragInteractionEnabled");
        }
        get isSurfaceInteractionDisabled() {
            return this.ae;
        }
        set isSurfaceInteractionDisabled(a) {
            this.ae = a;
            this.g("IsSurfaceInteractionDisabled");
        }
        get shouldDisplayMockData() {
            return this.af;
        }
        set shouldDisplayMockData(a) {
            this.af = a;
            this.g("ShouldDisplayMockData");
        }
        get shouldUseSkeletonStyleForMockData() {
            return this.ag;
        }
        set shouldUseSkeletonStyleForMockData(a) {
            this.ag = a;
            this.g("ShouldUseSkeletonStyleForMockData");
        }
        get legendEmptyValuesMode() {
            return this.bx;
        }
        set legendEmptyValuesMode(a) {
            this.bx = a;
            this.g("LegendEmptyValuesMode");
        }
        get formatLabelRef() {
            return this.bn;
        }
        set formatLabelRef(a) {
            this.bn = a;
            this.g("FormatLabelRef");
        }
        get formatLegendLabelRef() {
            return this.bo;
        }
        set formatLegendLabelRef(a) {
            this.bo = a;
            this.g("FormatLegendLabelRef");
        }
        get pixelScalingRatio() {
            return this.aw;
        }
        set pixelScalingRatio(a) {
            this.aw = a;
            this.g("PixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.an;
        }
        set actualPixelScalingRatio(a) {
            this.an = a;
            this.g("ActualPixelScalingRatio");
        }
        get selectionMode() {
            return this.cf;
        }
        set selectionMode(a) {
            this.cf = a;
            this.g("SelectionMode");
        }
        get selectedItem() {
            return this.bf;
        }
        set selectedItem(a) {
            this.bf = a;
            this.g("SelectedItem");
        }
        get selectedItems() {
            return this.q;
        }
        set selectedItems(a) {
            this.q = a;
            this.g("SelectedItems");
        }
        get textStyle() {
            return this.cl;
        }
        set textStyle(a) {
            this.cl = a;
            this.g("TextStyle");
        }
        get labelClickRef() {
            return this.bp;
        }
        set labelClickRef(a) {
            this.bp = a;
            this.g("LabelClickRef");
        }
        get selectedItemChangingRef() {
            return this.ca;
        }
        set selectedItemChangingRef(a) {
            this.ca = a;
            this.g("SelectedItemChangingRef");
        }
        get selectedItemsChangingRef() {
            return this.cc;
        }
        set selectedItemsChangingRef(a) {
            this.cc = a;
            this.g("SelectedItemsChangingRef");
        }
        get selectedItemChangedRef() {
            return this.b9;
        }
        set selectedItemChangedRef(a) {
            this.b9 = a;
            this.g("SelectedItemChangedRef");
        }
        get selectedItemsChangedRef() {
            return this.cb;
        }
        set selectedItemsChangedRef(a) {
            this.cb = a;
            this.g("SelectedItemsChangedRef");
        }
        get sliceClickRef() {
            return this.cg;
        }
        set sliceClickRef(a) {
            this.cg = a;
            this.g("SliceClickRef");
        }
        get sliceEnterRef() {
            return this.ch;
        }
        set sliceEnterRef(a) {
            this.ch = a;
            this.g("SliceEnterRef");
        }
        get sliceLeaveRef() {
            return this.cj;
        }
        set sliceLeaveRef(a) {
            this.cj = a;
            this.g("SliceLeaveRef");
        }
        get sliceHoverRef() {
            return this.ci;
        }
        set sliceHoverRef(a) {
            this.ci = a;
            this.g("SliceHoverRef");
        }
    }
    PieChartBaseDescription.$t = /*@__PURE__*/ markType(PieChartBaseDescription, 'PieChartBaseDescription', Description.$);
    return PieChartBaseDescription;
})();
