// igniteui-react-core/esm2015/lib/PlotAreaMouseEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let PlotAreaMouseEventArgsDescription = /*@__PURE__*/ (() => {
    class PlotAreaMouseEventArgsDescription extends Description {
        get_type() {
            return "PlotAreaMouseEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.i = null;
            this.n = false;
            this.l = null;
        }
        get chartPosition() {
            return this.h;
        }
        set chartPosition(a) {
            this.h = a;
            this.g("ChartPosition");
        }
        get plotAreaPosition() {
            return this.i;
        }
        set plotAreaPosition(a) {
            this.i = a;
            this.g("PlotAreaPosition");
        }
        get isDuringManipulation() {
            return this.n;
        }
        set isDuringManipulation(a) {
            this.n = a;
            this.g("IsDuringManipulation");
        }
        get viewer() {
            return this.l;
        }
        set viewer(a) {
            this.l = a;
            this.g("Viewer");
        }
    }
    PlotAreaMouseEventArgsDescription.$t = /*@__PURE__*/ markType(PlotAreaMouseEventArgsDescription, 'PlotAreaMouseEventArgsDescription', Description.$);
    return PlotAreaMouseEventArgsDescription;
})();
