// igniteui-react-core/esm2015/lib/PolarBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let PolarBaseDescription = /*@__PURE__*/ (() => {
    class PolarBaseDescription extends MarkerSeriesDescription {
        get_type() {
            return "PolarBase";
        }
        constructor() {
            super();
            this.gg = null;
            this.gp = null;
            this.gk = null;
            this.gj = null;
            this.gd = null;
            this.gm = null;
            this.fw = false;
            this.f4 = 0;
            this.gr = null;
            this.gq = null;
            this.gc = null;
            this.f1 = 0;
            this.fr = null;
            this.f5 = 0;
            this.f6 = 0;
            this.ft = false;
            this.gb = null;
            this.gl = null;
            this.f3 = 0;
            this.fv = false;
            this.fu = false;
            this.gn = null;
            this.ge = null;
            this.go = null;
            this.gf = null;
            this.gi = null;
            this.gh = null;
        }
        get angleMemberPath() {
            return this.gg;
        }
        set angleMemberPath(a) {
            this.gg = a;
            this.g("AngleMemberPath");
        }
        get radiusMemberPath() {
            return this.gp;
        }
        set radiusMemberPath(a) {
            this.gp = a;
            this.g("RadiusMemberPath");
        }
        get highlightedRadiusMemberPath() {
            return this.gk;
        }
        set highlightedRadiusMemberPath(a) {
            this.gk = a;
            this.g("HighlightedRadiusMemberPath");
        }
        get highlightedAngleMemberPath() {
            return this.gj;
        }
        set highlightedAngleMemberPath(a) {
            this.gj = a;
            this.g("HighlightedAngleMemberPath");
        }
        get angleAxisRef() {
            return this.gd;
        }
        set angleAxisRef(a) {
            this.gd = a;
            this.g("AngleAxisRef");
        }
        get radiusAxisRef() {
            return this.gm;
        }
        set radiusAxisRef(a) {
            this.gm = a;
            this.g("RadiusAxisRef");
        }
        get useCartesianInterpolation() {
            return this.fw;
        }
        set useCartesianInterpolation(a) {
            this.fw = a;
            this.g("UseCartesianInterpolation");
        }
        get maximumMarkers() {
            return this.f4;
        }
        set maximumMarkers(a) {
            this.f4 = a;
            this.g("MaximumMarkers");
        }
        get trendLineType() {
            return this.gr;
        }
        set trendLineType(a) {
            this.gr = a;
            this.g("TrendLineType");
        }
        get trendLineBrush() {
            return this.gq;
        }
        set trendLineBrush(a) {
            this.gq = a;
            this.g("TrendLineBrush");
        }
        get actualTrendLineBrush() {
            return this.gc;
        }
        set actualTrendLineBrush(a) {
            this.gc = a;
            this.g("ActualTrendLineBrush");
        }
        get trendLineThickness() {
            return this.f1;
        }
        set trendLineThickness(a) {
            this.f1 = a;
            this.g("TrendLineThickness");
        }
        get trendLineDashArray() {
            return this.fr;
        }
        set trendLineDashArray(a) {
            this.fr = a;
            this.g("TrendLineDashArray");
        }
        get trendLinePeriod() {
            return this.f5;
        }
        set trendLinePeriod(a) {
            this.f5 = a;
            this.g("TrendLinePeriod");
        }
        get trendLineZIndex() {
            return this.f6;
        }
        set trendLineZIndex(a) {
            this.f6 = a;
            this.g("TrendLineZIndex");
        }
        get clipSeriesToBounds() {
            return this.ft;
        }
        set clipSeriesToBounds(a) {
            this.ft = a;
            this.g("ClipSeriesToBounds");
        }
        get actualItemSearchMode() {
            return this.gb;
        }
        set actualItemSearchMode(a) {
            this.gb = a;
            this.g("ActualItemSearchMode");
        }
        get itemSearchMode() {
            return this.gl;
        }
        set itemSearchMode(a) {
            this.gl = a;
            this.g("ItemSearchMode");
        }
        get itemSearchThreshold() {
            return this.f3;
        }
        set itemSearchThreshold(a) {
            this.f3 = a;
            this.g("ItemSearchThreshold");
        }
        get isCustomPolarStyleAllowed() {
            return this.fv;
        }
        set isCustomPolarStyleAllowed(a) {
            this.fv = a;
            this.g("IsCustomPolarStyleAllowed");
        }
        get isCustomPolarMarkerStyleAllowed() {
            return this.fu;
        }
        set isCustomPolarMarkerStyleAllowed(a) {
            this.fu = a;
            this.g("IsCustomPolarMarkerStyleAllowed");
        }
        get radiusMemberAsLegendLabel() {
            return this.gn;
        }
        set radiusMemberAsLegendLabel(a) {
            this.gn = a;
            this.g("RadiusMemberAsLegendLabel");
        }
        get angleMemberAsLegendLabel() {
            return this.ge;
        }
        set angleMemberAsLegendLabel(a) {
            this.ge = a;
            this.g("AngleMemberAsLegendLabel");
        }
        get radiusMemberAsLegendUnit() {
            return this.go;
        }
        set radiusMemberAsLegendUnit(a) {
            this.go = a;
            this.g("RadiusMemberAsLegendUnit");
        }
        get angleMemberAsLegendUnit() {
            return this.gf;
        }
        set angleMemberAsLegendUnit(a) {
            this.gf = a;
            this.g("AngleMemberAsLegendUnit");
        }
        get assigningPolarStyleRef() {
            return this.gi;
        }
        set assigningPolarStyleRef(a) {
            this.gi = a;
            this.g("AssigningPolarStyleRef");
        }
        get assigningPolarMarkerStyleRef() {
            return this.gh;
        }
        set assigningPolarMarkerStyleRef(a) {
            this.gh = a;
            this.g("AssigningPolarMarkerStyleRef");
        }
    }
    PolarBaseDescription.$t = /*@__PURE__*/ markType(PolarBaseDescription, 'PolarBaseDescription', MarkerSeriesDescription.$);
    return PolarBaseDescription;
})();
