// igniteui-react-core/esm2015/lib/PopupVisualModelExport.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { StringBuilder } from "./StringBuilder";
/**
 * @hidden
 */
export let PopupVisualModelExport = /*@__PURE__*/ (() => {
    class PopupVisualModelExport extends Base {
        constructor() {
            super(...arguments);
            this.e = 0;
            this.d = false;
            this.c = null;
            this.b = null;
            this.a = null;
        }
        f() {
            let a = new StringBuilder(0);
            a.l("{");
            a.l("actualElevation : " + this.e + ", ");
            a.l("isShown : " + this.d + ", ");
            a.l("actualUmbraShadowColor : " + this.c.serialize() + ", ");
            a.l("actualPenumbraShadowColor : " + this.b.serialize() + ", ");
            a.l("actualAmbientShadowColor : " + this.a.serialize() + ", ");
            a.l("}");
            return a.toString();
        }
    }
    PopupVisualModelExport.$t = /*@__PURE__*/ markType(PopupVisualModelExport, 'PopupVisualModelExport');
    return PopupVisualModelExport;
})();
