// igniteui-react-core/esm2015/lib/Popup_OuterAnimationInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ValueType, markStruct } from "./type";
/**
 * @hidden
 */
export let Popup_OuterAnimationInfo = /*@__PURE__*/ (() => {
    class Popup_OuterAnimationInfo extends ValueType {
        constructor() {
            super();
            this.g = 0;
            this.h = 0;
            this.c = 0;
            this.f = 0;
            this.a = 0;
            this.d = 0;
            this.b = 0;
            this.e = 0;
        }
    }
    Popup_OuterAnimationInfo.$t = /*@__PURE__*/ markStruct(Popup_OuterAnimationInfo, 'Popup_OuterAnimationInfo');
    return Popup_OuterAnimationInfo;
})();
