// igniteui-react-core/esm2015/lib/PopupAlignment.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Indicates which side of the exclusion rectangle to which the popup is aligned.
 */
export var PopupAlignment = /*@__PURE__*/ (function (PopupAlignment) {
    /**
     * No specified direction.
     */
    PopupAlignment[PopupAlignment["Auto"] = 0] = "Auto";
    /**
     * Popup should be aligned to the near side. This usually means left or top.
     */
    PopupAlignment[PopupAlignment["Near"] = 1] = "Near";
    /**
     * Popup should be aligned to the near far. This usually means right or bottom.
     */
    PopupAlignment[PopupAlignment["Far"] = 2] = "Far";
    /**
     * Popup should be aligned in the middle.
     */
    PopupAlignment[PopupAlignment["Middle"] = 3] = "Middle";
    return PopupAlignment;
})({});
/**
 * @hidden
 */
export let PopupAlignment_$type = /*@__PURE__*/ markEnum('PopupAlignment', 'Auto,0|Near,1|Far,2|Middle,3');
