// igniteui-react-core/esm2015/lib/PropertyEditorDescriptionObjectDescription_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let PropertyEditorDescriptionObjectDescription = /*@__PURE__*/ (() => {
    class PropertyEditorDescriptionObjectDescription extends Description {
        get_type() {
            return "PropertyEditorDescriptionObject";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = null;
        }
        get descriptionType() {
            return this.j;
        }
        set descriptionType(a) {
            this.j = a;
            this.g("DescriptionType");
        }
        get properties() {
            return this.h;
        }
        set properties(a) {
            this.h = a;
            this.g("Properties");
        }
    }
    PropertyEditorDescriptionObjectDescription.$t = /*@__PURE__*/ markType(PropertyEditorDescriptionObjectDescription, 'PropertyEditorDescriptionObjectDescription', Description.$);
    return PropertyEditorDescriptionObjectDescription;
})();
/**
 * @hidden
 */
export let PropertyEditorPropertyDescriptionDescription = /*@__PURE__*/ (() => {
    class PropertyEditorPropertyDescriptionDescription extends Description {
        get_type() {
            return "PropertyEditorPropertyDescription";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.au = null;
            this.as = null;
            this.av = null;
            this.ax = null;
            this.ap = null;
            this.at = null;
            this.v = false;
            this.ar = null;
            this.j = null;
            this.k = null;
            this.ak = null;
            this.aj = null;
            this.ac = 0;
            this.z = 0;
            this.aa = 0;
            this.ab = 0;
            this.ad = 0;
            this.s = null;
            this.r = null;
            this.i = null;
            this.h = null;
            this.w = false;
            this.l = null;
            this.an = null;
            this.ao = null;
            this.aw = null;
            this.aq = null;
        }
        get propertyPath() {
            return this.au;
        }
        set propertyPath(a) {
            this.au = a;
            this.g("PropertyPath");
        }
        get label() {
            return this.as;
        }
        set label(a) {
            this.as = a;
            this.g("Label");
        }
        get subtitle() {
            return this.av;
        }
        set subtitle(a) {
            this.av = a;
            this.g("Subtitle");
        }
        get valueType() {
            return this.ax;
        }
        set valueType(a) {
            this.ax = a;
            this.g("ValueType");
        }
        get coercedValueType() {
            return this.ap;
        }
        set coercedValueType(a) {
            this.ap = a;
            this.g("CoercedValueType");
        }
        get propertyDescriptionType() {
            return this.at;
        }
        set propertyDescriptionType(a) {
            this.at = a;
            this.g("PropertyDescriptionType");
        }
        get shouldOverrideDefaultEditor() {
            return this.v;
        }
        set shouldOverrideDefaultEditor(a) {
            this.v = a;
            this.g("ShouldOverrideDefaultEditor");
        }
        get elementDescriptionType() {
            return this.ar;
        }
        set elementDescriptionType(a) {
            this.ar = a;
            this.g("ElementDescriptionType");
        }
        get dropDownNames() {
            return this.j;
        }
        set dropDownNames(a) {
            this.j = a;
            this.g("DropDownNames");
        }
        get dropDownValues() {
            return this.k;
        }
        set dropDownValues(a) {
            this.k = a;
            this.g("DropDownValues");
        }
        get primitiveValue() {
            return this.ak;
        }
        set primitiveValue(a) {
            this.ak = a;
            this.g("PrimitiveValue");
        }
        get coercedPrimitiveValue() {
            return this.aj;
        }
        set coercedPrimitiveValue(a) {
            this.aj = a;
            this.g("CoercedPrimitiveValue");
        }
        get min() {
            return this.ac;
        }
        set min(a) {
            this.ac = a;
            this.g("Min");
        }
        get editorWidth() {
            return this.z;
        }
        set editorWidth(a) {
            this.z = a;
            this.g("EditorWidth");
        }
        get labelWidth() {
            return this.aa;
        }
        set labelWidth(a) {
            this.aa = a;
            this.g("LabelWidth");
        }
        get max() {
            return this.ab;
        }
        set max(a) {
            this.ab = a;
            this.g("Max");
        }
        get step() {
            return this.ad;
        }
        set step(a) {
            this.ad = a;
            this.g("Step");
        }
        get complexValue() {
            return this.s;
        }
        set complexValue(a) {
            this.s = a;
            this.g("ComplexValue");
        }
        get coercedComplexValue() {
            return this.r;
        }
        set coercedComplexValue(a) {
            this.r = a;
            this.g("CoercedComplexValue");
        }
        get complexValues() {
            return this.i;
        }
        set complexValues(a) {
            this.i = a;
            this.g("ComplexValues");
        }
        get coercedComplexValues() {
            return this.h;
        }
        set coercedComplexValues(a) {
            this.h = a;
            this.g("CoercedComplexValues");
        }
        get useCoercedValue() {
            return this.w;
        }
        set useCoercedValue(a) {
            this.w = a;
            this.g("UseCoercedValue");
        }
        get properties() {
            return this.l;
        }
        set properties(a) {
            this.l = a;
            this.g("Properties");
        }
        get buttonClickedRef() {
            return this.an;
        }
        set buttonClickedRef(a) {
            this.an = a;
            this.g("ButtonClickedRef");
        }
        get changedRef() {
            return this.ao;
        }
        set changedRef(a) {
            this.ao = a;
            this.g("ChangedRef");
        }
        get targetPropertyUpdatingRef() {
            return this.aw;
        }
        set targetPropertyUpdatingRef(a) {
            this.aw = a;
            this.g("TargetPropertyUpdatingRef");
        }
        get coercingValueRef() {
            return this.aq;
        }
        set coercingValueRef(a) {
            this.aq = a;
            this.g("CoercingValueRef");
        }
    }
    PropertyEditorPropertyDescriptionDescription.$t = /*@__PURE__*/ markType(PropertyEditorPropertyDescriptionDescription, 'PropertyEditorPropertyDescriptionDescription', Description.$);
    return PropertyEditorPropertyDescriptionDescription;
})();
