// igniteui-react-core/esm2015/lib/ProportionalCategoryAngleAxisDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategoryAngleAxisDescription } from "./CategoryAngleAxisDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ProportionalCategoryAngleAxisDescription = /*@__PURE__*/ (() => {
    class ProportionalCategoryAngleAxisDescription extends CategoryAngleAxisDescription {
        get_type() {
            return "ProportionalCategoryAngleAxis";
        }
        constructor() {
            super();
            this.el = null;
            this.eh = 0;
            this.ek = null;
            this.ej = null;
        }
        get valueMemberPath() {
            return this.el;
        }
        set valueMemberPath(a) {
            this.el = a;
            this.g("ValueMemberPath");
        }
        get othersCategoryThreshold() {
            return this.eh;
        }
        set othersCategoryThreshold(a) {
            this.eh = a;
            this.g("OthersCategoryThreshold");
        }
        get othersCategoryType() {
            return this.ek;
        }
        set othersCategoryType(a) {
            this.ek = a;
            this.g("OthersCategoryType");
        }
        get othersCategoryText() {
            return this.ej;
        }
        set othersCategoryText(a) {
            this.ej = a;
            this.g("OthersCategoryText");
        }
    }
    ProportionalCategoryAngleAxisDescription.$t = /*@__PURE__*/ markType(ProportionalCategoryAngleAxisDescription, 'ProportionalCategoryAngleAxisDescription', CategoryAngleAxisDescription.$);
    return ProportionalCategoryAngleAxisDescription;
})();
