// igniteui-react-core/esm2015/lib/PropertyEditorPanelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let PropertyEditorPanelDescription = /*@__PURE__*/ (() => {
    class PropertyEditorPanelDescription extends Description {
        get_type() {
            return "PropertyEditorPanel";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.i = null;
            this.v = 0;
            this.ai = null;
            this.r = 0;
            this.z = null;
            this.s = 0;
            this.aa = null;
            this.ab = null;
            this.ad = null;
            this.ac = null;
            this.ag = null;
            this.ae = null;
            this.l = false;
            this.n = false;
            this.m = false;
            this.y = null;
            this.ah = null;
            this.aj = null;
            this.af = null;
            this.x = null;
        }
        get actualDataSource() {
            return this.h;
        }
        set actualDataSource(a) {
            this.h = a;
            this.g("ActualDataSource");
        }
        get properties() {
            return this.i;
        }
        set properties(a) {
            this.i = a;
            this.g("Properties");
        }
        get rowHeight() {
            return this.v;
        }
        set rowHeight(a) {
            this.v = a;
            this.g("RowHeight");
        }
        get updateMode() {
            return this.ai;
        }
        set updateMode(a) {
            this.ai = a;
            this.g("UpdateMode");
        }
        get actualRowHeight() {
            return this.r;
        }
        set actualRowHeight(a) {
            this.r = a;
            this.g("ActualRowHeight");
        }
        get cellFontFamily() {
            return this.z;
        }
        set cellFontFamily(a) {
            this.z = a;
            this.g("CellFontFamily");
        }
        get cellFontSize() {
            return this.s;
        }
        set cellFontSize(a) {
            this.s = a;
            this.g("CellFontSize");
        }
        get cellFontStyle() {
            return this.aa;
        }
        set cellFontStyle(a) {
            this.aa = a;
            this.g("CellFontStyle");
        }
        get cellFontWeight() {
            return this.ab;
        }
        set cellFontWeight(a) {
            this.ab = a;
            this.g("CellFontWeight");
        }
        get descriptionContextRef() {
            return this.ad;
        }
        set descriptionContextRef(a) {
            this.ad = a;
            this.g("DescriptionContextRef");
        }
        get componentRendererRef() {
            return this.ac;
        }
        set componentRendererRef(a) {
            this.ac = a;
            this.g("ComponentRendererRef");
        }
        get targetRef() {
            return this.ag;
        }
        set targetRef(a) {
            this.ag = a;
            this.g("TargetRef");
        }
        get descriptionType() {
            return this.ae;
        }
        set descriptionType(a) {
            this.ae = a;
            this.g("DescriptionType");
        }
        get isHorizontal() {
            return this.l;
        }
        set isHorizontal(a) {
            this.l = a;
            this.g("IsHorizontal");
        }
        get isWrappingEnabled() {
            return this.n;
        }
        set isWrappingEnabled(a) {
            this.n = a;
            this.g("IsWrappingEnabled");
        }
        get isIndirectModeEnabled() {
            return this.m;
        }
        set isIndirectModeEnabled(a) {
            this.m = a;
            this.g("IsIndirectModeEnabled");
        }
        get backgroundColor() {
            return this.y;
        }
        set backgroundColor(a) {
            this.y = a;
            this.g("BackgroundColor");
        }
        get textColor() {
            return this.ah;
        }
        set textColor(a) {
            this.ah = a;
            this.g("TextColor");
        }
        get width() {
            return this.aj;
        }
        set width(a) {
            this.aj = a;
            this.g("Width");
        }
        get height() {
            return this.af;
        }
        set height(a) {
            this.af = a;
            this.g("Height");
        }
        get background() {
            return this.x;
        }
        set background(a) {
            this.x = a;
            this.g("Background");
        }
    }
    PropertyEditorPanelDescription.$t = /*@__PURE__*/ markType(PropertyEditorPanelDescription, 'PropertyEditorPanelDescription', Description.$);
    return PropertyEditorPanelDescription;
})();
