// igniteui-react-core/esm2015/lib/PropertyEditorPropertyDescriptionCoercingValueEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let PropertyEditorPropertyDescriptionCoercingValueEventArgsDescription = /*@__PURE__*/ (() => {
    class PropertyEditorPropertyDescriptionCoercingValueEventArgsDescription extends Description {
        get_type() {
            return "PropertyEditorPropertyDescriptionCoercingValueEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.j = null;
            this.h = null;
        }
        get value() {
            return this.l;
        }
        set value(a) {
            this.l = a;
            this.g("Value");
        }
        get complexValue() {
            return this.j;
        }
        set complexValue(a) {
            this.j = a;
            this.g("ComplexValue");
        }
        get complexValues() {
            return this.h;
        }
        set complexValues(a) {
            this.h = a;
            this.g("ComplexValues");
        }
    }
    PropertyEditorPropertyDescriptionCoercingValueEventArgsDescription.$t = /*@__PURE__*/ markType(PropertyEditorPropertyDescriptionCoercingValueEventArgsDescription, 'PropertyEditorPropertyDescriptionCoercingValueEventArgsDescription', Description.$);
    return PropertyEditorPropertyDescriptionCoercingValueEventArgsDescription;
})();
