// igniteui-react-core/esm2015/lib/RadialLabelMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Describes available radial label modes.
 */
export var RadialLabelMode = /*@__PURE__*/ (function (RadialLabelMode) {
    /**
     * Labelling mode is determined automatically.
     */
    RadialLabelMode[RadialLabelMode["Auto"] = 0] = "Auto";
    /**
     * Labelling mode is defined by the settings of the callout layer.
     */
    RadialLabelMode[RadialLabelMode["Normal"] = 1] = "Normal";
    /**
     * The value of the item is displayed.
     */
    RadialLabelMode[RadialLabelMode["Value"] = 2] = "Value";
    /**
     * The percentage of the whole is displayed, if using a proportional axis, otherwise the value is used as a fallback.
     */
    RadialLabelMode[RadialLabelMode["Percentage"] = 3] = "Percentage";
    /**
     * The value and the percentage are displayed, if using a proportional axis, otherwise the value is used as a fallback.
     */
    RadialLabelMode[RadialLabelMode["ValueAndPercentage"] = 4] = "ValueAndPercentage";
    /**
     * The label of the category is used as a label.
     */
    RadialLabelMode[RadialLabelMode["Label"] = 5] = "Label";
    /**
     * The label of the category and the value are used as the label.
     */
    RadialLabelMode[RadialLabelMode["LabelAndValue"] = 6] = "LabelAndValue";
    /**
     * The label of the category and the percentage are used as the label, if using a proportional axis, otherwise the label and value is used as a fallback.
     */
    RadialLabelMode[RadialLabelMode["LabelAndPercentage"] = 7] = "LabelAndPercentage";
    /**
     * The label of the category and the value and the percentage are used as the label, if using a proportional axis, otherwise the label and value is used as a fallback.
     */
    RadialLabelMode[RadialLabelMode["LabelAndValueAndPercentage"] = 8] = "LabelAndValueAndPercentage";
    return RadialLabelMode;
})({});
/**
 * @hidden
 */
export let RadialLabelMode_$type = /*@__PURE__*/ markEnum('RadialLabelMode', 'Auto,0|Normal,1|Value,2|Percentage,3|ValueAndPercentage,4|Label,5|LabelAndValue,6|LabelAndPercentage,7|LabelAndValueAndPercentage,8');
