// igniteui-react-core/esm2015/lib/RadialPieSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnchoredRadialSeriesDescription } from "./AnchoredRadialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let RadialPieSeriesDescription = /*@__PURE__*/ (() => {
    class RadialPieSeriesDescription extends AnchoredRadialSeriesDescription {
        get_type() {
            return "RadialPieSeries";
        }
        constructor() {
            super();
            this.ht = 0;
            this.hu = 0;
            this.hy = null;
            this.hx = null;
        }
        get radiusX() {
            return this.ht;
        }
        set radiusX(a) {
            this.ht = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.hu;
        }
        set radiusY(a) {
            this.hu = a;
            this.g("RadiusY");
        }
        get legendLabelMemberPath() {
            return this.hy;
        }
        set legendLabelMemberPath(a) {
            this.hy = a;
            this.g("LegendLabelMemberPath");
        }
        get legendEmptyValuesMode() {
            return this.hx;
        }
        set legendEmptyValuesMode(a) {
            this.hx = a;
            this.g("LegendEmptyValuesMode");
        }
    }
    RadialPieSeriesDescription.$t = /*@__PURE__*/ markType(RadialPieSeriesDescription, 'RadialPieSeriesDescription', AnchoredRadialSeriesDescription.$);
    return RadialPieSeriesDescription;
})();
