// igniteui-react-core/esm2015/lib/RangeCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let RangeCategorySeriesDescription = /*@__PURE__*/ (() => {
    class RangeCategorySeriesDescription extends CategorySeriesDescription {
        get_type() {
            return "RangeCategorySeries";
        }
        constructor() {
            super();
            this.ga = null;
            this.f9 = null;
            this.f8 = null;
            this.f7 = null;
        }
        get lowMemberPath() {
            return this.ga;
        }
        set lowMemberPath(a) {
            this.ga = a;
            this.g("LowMemberPath");
        }
        get highMemberPath() {
            return this.f9;
        }
        set highMemberPath(a) {
            this.f9 = a;
            this.g("HighMemberPath");
        }
        get highlightedLowMemberPath() {
            return this.f8;
        }
        set highlightedLowMemberPath(a) {
            this.f8 = a;
            this.g("HighlightedLowMemberPath");
        }
        get highlightedHighMemberPath() {
            return this.f7;
        }
        set highlightedHighMemberPath(a) {
            this.f7 = a;
            this.g("HighlightedHighMemberPath");
        }
    }
    RangeCategorySeriesDescription.$t = /*@__PURE__*/ markType(RangeCategorySeriesDescription, 'RangeCategorySeriesDescription', CategorySeriesDescription.$);
    return RangeCategorySeriesDescription;
})();
