// igniteui-react-core/esm2015/lib/RectangleGeometry.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Geometry } from "./Geometry";
import { markType } from "./type";
/**
 * @hidden
 */
export let RectangleGeometry = /*@__PURE__*/ (() => {
    class RectangleGeometry extends Geometry {
        constructor() {
            super(...arguments);
            this.e = null;
            this.c = 0;
            this.d = 0;
        }
        get_b() {
            return 2;
        }
        get b() {
            return this.get_b();
        }
    }
    RectangleGeometry.$t = /*@__PURE__*/ markType(RectangleGeometry, 'RectangleGeometry', Geometry.$);
    return RectangleGeometry;
})();
