// igniteui-react-core/esm2015/lib/Rectangle.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Shape } from "./Shape";
import { Rect } from "./Rect";
import { markType } from "./type";
/**
 * @hidden
 */
export let Rectangle = /*@__PURE__*/ (() => {
    class Rectangle extends Shape {
        get ap() {
            return this.an;
        }
        set ap(a) {
            this.an = a;
        }
        get aq() {
            return this.ao;
        }
        set aq(a) {
            this.ao = a;
        }
        constructor() {
            super();
            this.as = null;
            this.an = 0;
            this.ao = 0;
            this.as = new Rect(0, 0, 0, 0, 0);
        }
        ar(a) {
        }
    }
    Rectangle.$t = /*@__PURE__*/ markType(Rectangle, 'Rectangle', Shape.$);
    return Rectangle;
})();
