// igniteui-react-core/esm2015/lib/ResponsiveStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ResponsiveStateDescription = /*@__PURE__*/ (() => {
    class ResponsiveStateDescription extends Description {
        get_type() {
            return "ResponsiveState";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.k = 0;
            this.j = 0;
            this.h = false;
            this.p = null;
            this.o = null;
            this.q = null;
        }
        get name() {
            return this.n;
        }
        set name(a) {
            this.n = a;
            this.g("Name");
        }
        get minimumWidth() {
            return this.k;
        }
        set minimumWidth(a) {
            this.k = a;
            this.g("MinimumWidth");
        }
        get maximumWidth() {
            return this.j;
        }
        set maximumWidth(a) {
            this.j = a;
            this.g("MaximumWidth");
        }
        get isManualState() {
            return this.h;
        }
        set isManualState(a) {
            this.h = a;
            this.g("IsManualState");
        }
        get stateEnteringRef() {
            return this.p;
        }
        set stateEnteringRef(a) {
            this.p = a;
            this.g("StateEnteringRef");
        }
        get stateEnteredRef() {
            return this.o;
        }
        set stateEnteredRef(a) {
            this.o = a;
            this.g("StateEnteredRef");
        }
        get stateExitedRef() {
            return this.q;
        }
        set stateExitedRef(a) {
            this.q = a;
            this.g("StateExitedRef");
        }
    }
    ResponsiveStateDescription.$t = /*@__PURE__*/ markType(ResponsiveStateDescription, 'ResponsiveStateDescription', Description.$);
    return ResponsiveStateDescription;
})();
