// igniteui-react-core/esm2015/lib/ResponsivePhaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ResponsivePhaseDescription = /*@__PURE__*/ (() => {
    class ResponsivePhaseDescription extends Description {
        get_type() {
            return "ResponsivePhase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = 0;
        }
        get name() {
            return this.j;
        }
        set name(a) {
            this.j = a;
            this.g("Name");
        }
        get delayMilliseconds() {
            return this.h;
        }
        set delayMilliseconds(a) {
            this.h = a;
            this.g("DelayMilliseconds");
        }
    }
    ResponsivePhaseDescription.$t = /*@__PURE__*/ markType(ResponsivePhaseDescription, 'ResponsivePhaseDescription', Description.$);
    return ResponsivePhaseDescription;
})();
