// igniteui-react-core/esm2015/lib/RingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let RingDescription = /*@__PURE__*/ (() => {
    class RingDescription extends Description {
        get_type() {
            return "Ring";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.r = 0;
            this.n = 0;
            this.l = null;
            this.h = null;
            this.j = null;
            this.o = 0;
        }
        get index() {
            return this.r;
        }
        set index(a) {
            this.r = a;
            this.g("Index");
        }
        get innerExtend() {
            return this.n;
        }
        set innerExtend(a) {
            this.n = a;
            this.g("InnerExtend");
        }
        get controlSize() {
            return this.l;
        }
        set controlSize(a) {
            this.l = a;
            this.g("ControlSize");
        }
        get center() {
            return this.h;
        }
        set center(a) {
            this.h = a;
            this.g("Center");
        }
        get ringSeries() {
            return this.j;
        }
        set ringSeries(a) {
            this.j = a;
            this.g("RingSeries");
        }
        get ringBreadth() {
            return this.o;
        }
        set ringBreadth(a) {
            this.o = a;
            this.g("RingBreadth");
        }
    }
    RingDescription.$t = /*@__PURE__*/ markType(RingDescription, 'RingDescription', Description.$);
    return RingDescription;
})();
