// igniteui-react-core/esm2015/lib/ScatterPolygonSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ShapeSeriesBaseDescription } from "./ShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ScatterPolygonSeriesDescription = /*@__PURE__*/ (() => {
    class ScatterPolygonSeriesDescription extends ShapeSeriesBaseDescription {
        get_type() {
            return "ScatterPolygonSeries";
        }
        constructor() {
            super();
            this.f1 = null;
            this.ge = null;
            this.gd = null;
            this.f3 = 0;
            this.f7 = null;
            this.f8 = null;
            this.f5 = null;
            this.gb = null;
            this.f6 = null;
            this.f9 = null;
            this.gc = null;
            this.ga = null;
        }
        get shapeStyle() {
            return this.f1;
        }
        set shapeStyle(a) {
            this.f1 = a;
            this.g("ShapeStyle");
        }
        get markerType() {
            return this.ge;
        }
        set markerType(a) {
            this.ge = a;
            this.g("MarkerType");
        }
        get markerTemplateRef() {
            return this.gd;
        }
        set markerTemplateRef(a) {
            this.gd = a;
            this.g("MarkerTemplateRef");
        }
        get markerThickness() {
            return this.f3;
        }
        set markerThickness(a) {
            this.f3 = a;
            this.g("MarkerThickness");
        }
        get actualMarkerTemplateRef() {
            return this.f7;
        }
        set actualMarkerTemplateRef(a) {
            this.f7 = a;
            this.g("ActualMarkerTemplateRef");
        }
        get markerBrush() {
            return this.f8;
        }
        set markerBrush(a) {
            this.f8 = a;
            this.g("MarkerBrush");
        }
        get actualMarkerBrush() {
            return this.f5;
        }
        set actualMarkerBrush(a) {
            this.f5 = a;
            this.g("ActualMarkerBrush");
        }
        get markerOutline() {
            return this.gb;
        }
        set markerOutline(a) {
            this.gb = a;
            this.g("MarkerOutline");
        }
        get actualMarkerOutline() {
            return this.f6;
        }
        set actualMarkerOutline(a) {
            this.f6 = a;
            this.g("ActualMarkerOutline");
        }
        get markerCollisionAvoidance() {
            return this.f9;
        }
        set markerCollisionAvoidance(a) {
            this.f9 = a;
            this.g("MarkerCollisionAvoidance");
        }
        get markerOutlineMode() {
            return this.gc;
        }
        set markerOutlineMode(a) {
            this.gc = a;
            this.g("MarkerOutlineMode");
        }
        get markerFillMode() {
            return this.ga;
        }
        set markerFillMode(a) {
            this.ga = a;
            this.g("MarkerFillMode");
        }
    }
    ScatterPolygonSeriesDescription.$t = /*@__PURE__*/ markType(ScatterPolygonSeriesDescription, 'ScatterPolygonSeriesDescription', ShapeSeriesBaseDescription.$);
    return ScatterPolygonSeriesDescription;
})();
