// igniteui-react-core/esm2015/lib/ScatterContourSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterTriangulationSeriesDescription } from "./ScatterTriangulationSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ScatterContourSeriesDescription = /*@__PURE__*/ (() => {
    class ScatterContourSeriesDescription extends ScatterTriangulationSeriesDescription {
        get_type() {
            return "ScatterContourSeries";
        }
        constructor() {
            super();
            this.f1 = null;
            this.fz = null;
            this.f0 = null;
            this.fw = null;
            this.fv = null;
            this.ft = null;
        }
        get valueMemberPath() {
            return this.f1;
        }
        set valueMemberPath(a) {
            this.f1 = a;
            this.g("ValueMemberPath");
        }
        get valueMemberAsLegendLabel() {
            return this.fz;
        }
        set valueMemberAsLegendLabel(a) {
            this.fz = a;
            this.g("ValueMemberAsLegendLabel");
        }
        get valueMemberAsLegendUnit() {
            return this.f0;
        }
        set valueMemberAsLegendUnit(a) {
            this.f0 = a;
            this.g("ValueMemberAsLegendUnit");
        }
        get fillScale() {
            return this.fw;
        }
        set fillScale(a) {
            this.fw = a;
            this.g("FillScale");
        }
        get actualFillScale() {
            return this.fv;
        }
        set actualFillScale(a) {
            this.fv = a;
            this.g("ActualFillScale");
        }
        get valueResolver() {
            return this.ft;
        }
        set valueResolver(a) {
            this.ft = a;
            this.g("ValueResolver");
        }
    }
    ScatterContourSeriesDescription.$t = /*@__PURE__*/ markType(ScatterContourSeriesDescription, 'ScatterContourSeriesDescription', ScatterTriangulationSeriesDescription.$);
    return ScatterContourSeriesDescription;
})();
