// igniteui-react-core/esm2015/lib/ScatterTriangulationSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ScatterTriangulationSeriesDescription = /*@__PURE__*/ (() => {
    class ScatterTriangulationSeriesDescription extends SeriesDescription {
        get_type() {
            return "ScatterTriangulationSeries";
        }
        constructor() {
            super();
            this.fb = null;
            this.ff = null;
            this.e8 = null;
            this.fc = null;
            this.e3 = null;
            this.e4 = null;
            this.e5 = null;
            this.e6 = null;
            this.e9 = null;
            this.fd = null;
            this.fa = null;
            this.fe = null;
            this.e7 = null;
        }
        get xMemberPath() {
            return this.fb;
        }
        set xMemberPath(a) {
            this.fb = a;
            this.g("XMemberPath");
        }
        get yMemberPath() {
            return this.ff;
        }
        set yMemberPath(a) {
            this.ff = a;
            this.g("YMemberPath");
        }
        get xAxisRef() {
            return this.e8;
        }
        set xAxisRef(a) {
            this.e8 = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.fc;
        }
        set yAxisRef(a) {
            this.fc = a;
            this.g("YAxisRef");
        }
        get trianglesSourceRef() {
            return this.e3;
        }
        set trianglesSourceRef(a) {
            this.e3 = a;
            this.g("TrianglesSourceRef");
        }
        get triangleVertexMemberPath1() {
            return this.e4;
        }
        set triangleVertexMemberPath1(a) {
            this.e4 = a;
            this.g("TriangleVertexMemberPath1");
        }
        get triangleVertexMemberPath2() {
            return this.e5;
        }
        set triangleVertexMemberPath2(a) {
            this.e5 = a;
            this.g("TriangleVertexMemberPath2");
        }
        get triangleVertexMemberPath3() {
            return this.e6;
        }
        set triangleVertexMemberPath3(a) {
            this.e6 = a;
            this.g("TriangleVertexMemberPath3");
        }
        get xMemberAsLegendLabel() {
            return this.e9;
        }
        set xMemberAsLegendLabel(a) {
            this.e9 = a;
            this.g("XMemberAsLegendLabel");
        }
        get yMemberAsLegendLabel() {
            return this.fd;
        }
        set yMemberAsLegendLabel(a) {
            this.fd = a;
            this.g("YMemberAsLegendLabel");
        }
        get xMemberAsLegendUnit() {
            return this.fa;
        }
        set xMemberAsLegendUnit(a) {
            this.fa = a;
            this.g("XMemberAsLegendUnit");
        }
        get yMemberAsLegendUnit() {
            return this.fe;
        }
        set yMemberAsLegendUnit(a) {
            this.fe = a;
            this.g("YMemberAsLegendUnit");
        }
        get triangulationStatusChangedRef() {
            return this.e7;
        }
        set triangulationStatusChangedRef(a) {
            this.e7 = a;
            this.g("TriangulationStatusChangedRef");
        }
    }
    ScatterTriangulationSeriesDescription.$t = /*@__PURE__*/ markType(ScatterTriangulationSeriesDescription, 'ScatterTriangulationSeriesDescription', SeriesDescription.$);
    return ScatterTriangulationSeriesDescription;
})();
