// igniteui-react-core/esm2015/lib/ScatterAreaSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterTriangulationSeriesDescription } from "./ScatterTriangulationSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ScatterAreaSeriesDescription = /*@__PURE__*/ (() => {
    class ScatterAreaSeriesDescription extends ScatterTriangulationSeriesDescription {
        get_type() {
            return "ScatterAreaSeries";
        }
        constructor() {
            super();
            this.fu = null;
            this.fz = null;
            this.ft = null;
            this.fx = null;
            this.fy = null;
        }
        get colorScale() {
            return this.fu;
        }
        set colorScale(a) {
            this.fu = a;
            this.g("ColorScale");
        }
        get colorMemberPath() {
            return this.fz;
        }
        set colorMemberPath(a) {
            this.fz = a;
            this.g("ColorMemberPath");
        }
        get actualColorScale() {
            return this.ft;
        }
        set actualColorScale(a) {
            this.ft = a;
            this.g("ActualColorScale");
        }
        get colorMemberAsLegendLabel() {
            return this.fx;
        }
        set colorMemberAsLegendLabel(a) {
            this.fx = a;
            this.g("ColorMemberAsLegendLabel");
        }
        get colorMemberAsLegendUnit() {
            return this.fy;
        }
        set colorMemberAsLegendUnit(a) {
            this.fy = a;
            this.g("ColorMemberAsLegendUnit");
        }
    }
    ScatterAreaSeriesDescription.$t = /*@__PURE__*/ markType(ScatterAreaSeriesDescription, 'ScatterAreaSeriesDescription', ScatterTriangulationSeriesDescription.$);
    return ScatterAreaSeriesDescription;
})();
