// igniteui-react-core/esm2015/lib/SectionHeaderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let SectionHeaderDescription = /*@__PURE__*/ (() => {
    class SectionHeaderDescription extends DefinitionBaseDescription {
        get_type() {
            return "SectionHeader";
        }
        constructor() {
            super();
            this.cz = null;
            this.cy = null;
            this.cr = 0;
            this.ct = 0;
            this.cs = 0;
            this.cq = 0;
            this.co = false;
            this.c0 = null;
        }
        get selectedBackground() {
            return this.cz;
        }
        set selectedBackground(a) {
            this.cz = a;
            this.g("SelectedBackground");
        }
        get actualSelectedBackground() {
            return this.cy;
        }
        set actualSelectedBackground(a) {
            this.cy = a;
            this.g("ActualSelectedBackground");
        }
        get paddingLeft() {
            return this.cr;
        }
        set paddingLeft(a) {
            this.cr = a;
            this.g("PaddingLeft");
        }
        get paddingTop() {
            return this.ct;
        }
        set paddingTop(a) {
            this.ct = a;
            this.g("PaddingTop");
        }
        get paddingRight() {
            return this.cs;
        }
        set paddingRight(a) {
            this.cs = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.cq;
        }
        set paddingBottom(a) {
            this.cq = a;
            this.g("PaddingBottom");
        }
        get isCollapsable() {
            return this.co;
        }
        set isCollapsable(a) {
            this.co = a;
            this.g("IsCollapsable");
        }
        get summaryDisplayMode() {
            return this.c0;
        }
        set summaryDisplayMode(a) {
            this.c0 = a;
            this.g("SummaryDisplayMode");
        }
    }
    SectionHeaderDescription.$t = /*@__PURE__*/ markType(SectionHeaderDescription, 'SectionHeaderDescription', DefinitionBaseDescription.$);
    return SectionHeaderDescription;
})();
