// igniteui-react-core/esm2015/lib/SectionHeaderCellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CellInfoDescription } from "./CellInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let SectionHeaderCellInfoDescription = /*@__PURE__*/ (() => {
    class SectionHeaderCellInfoDescription extends CellInfoDescription {
        get_type() {
            return "SectionHeaderCellInfo";
        }
        constructor() {
            super();
            this.fo = null;
            this.fn = null;
            this.fp = null;
        }
        get resolvedText() {
            return this.fo;
        }
        set resolvedText(a) {
            this.fo = a;
            this.g("ResolvedText");
        }
        get resolvedSummaryText() {
            return this.fn;
        }
        set resolvedSummaryText(a) {
            this.fn = a;
            this.g("ResolvedSummaryText");
        }
        get summaryDisplayMode() {
            return this.fp;
        }
        set summaryDisplayMode(a) {
            this.fp = a;
            this.g("SummaryDisplayMode");
        }
    }
    SectionHeaderCellInfoDescription.$t = /*@__PURE__*/ markType(SectionHeaderCellInfoDescription, 'SectionHeaderCellInfoDescription', CellInfoDescription.$);
    SectionHeaderCellInfoDescription.__marshalByValue1 = true;
    SectionHeaderCellInfoDescription.__marshalByValueAlias1 = "SectionHeaderCellModel";
    return SectionHeaderCellInfoDescription;
})();
