// igniteui-react-core/esm2015/lib/SelectedItemsChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SelectedItemsChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class SelectedItemsChangedEventArgsDescription extends Description {
        get_type() {
            return "SelectedItemsChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.i = null;
            this.h = null;
        }
        get oldItems() {
            return this.j;
        }
        set oldItems(a) {
            this.j = a;
            this.g("OldItems");
        }
        get newItems() {
            return this.i;
        }
        set newItems(a) {
            this.i = a;
            this.g("NewItems");
        }
        get currentItems() {
            return this.h;
        }
        set currentItems(a) {
            this.h = a;
            this.g("CurrentItems");
        }
    }
    SelectedItemsChangedEventArgsDescription.$t = /*@__PURE__*/ markType(SelectedItemsChangedEventArgsDescription, 'SelectedItemsChangedEventArgsDescription', Description.$);
    return SelectedItemsChangedEventArgsDescription;
})();
