// igniteui-react-core/esm2015/lib/SelectedValueChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SelectedValueChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class SelectedValueChangedEventArgsDescription extends Description {
        get_type() {
            return "SelectedValueChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = new Date();
            this.h = new Date();
        }
        get oldValue() {
            return this.i;
        }
        set oldValue(a) {
            this.i = a;
            this.g("OldValue");
        }
        get newValue() {
            return this.h;
        }
        set newValue(a) {
            this.h = a;
            this.g("NewValue");
        }
    }
    SelectedValueChangedEventArgsDescription.$t = /*@__PURE__*/ markType(SelectedValueChangedEventArgsDescription, 'SelectedValueChangedEventArgsDescription', Description.$);
    return SelectedValueChangedEventArgsDescription;
})();
