// igniteui-react-core/esm2015/lib/SeriesLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SeriesLayerDescription = /*@__PURE__*/ (() => {
    class SeriesLayerDescription extends Description {
        get_type() {
            return "SeriesLayer";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = 0;
            this.h = null;
        }
        get zIndex() {
            return this.j;
        }
        set zIndex(a) {
            this.j = a;
            this.g("ZIndex");
        }
        get propertyOverlays() {
            return this.h;
        }
        set propertyOverlays(a) {
            this.h = a;
            this.g("PropertyOverlays");
        }
    }
    SeriesLayerDescription.$t = /*@__PURE__*/ markType(SeriesLayerDescription, 'SeriesLayerDescription', Description.$);
    return SeriesLayerDescription;
})();
