// igniteui-react-core/esm2015/lib/SeriesViewerManipulationEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SeriesViewerManipulationEventArgsDescription = /*@__PURE__*/ (() => {
    class SeriesViewerManipulationEventArgsDescription extends Description {
        get_type() {
            return "SeriesViewerManipulationEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = false;
            this.i = false;
            this.h = false;
        }
        get isZoomPan() {
            return this.j;
        }
        set isZoomPan(a) {
            this.j = a;
            this.g("IsZoomPan");
        }
        get isDragZoom() {
            return this.i;
        }
        set isDragZoom(a) {
            this.i = a;
            this.g("IsDragZoom");
        }
        get isDragSelect() {
            return this.h;
        }
        set isDragSelect(a) {
            this.h = a;
            this.g("IsDragSelect");
        }
    }
    SeriesViewerManipulationEventArgsDescription.$t = /*@__PURE__*/ markType(SeriesViewerManipulationEventArgsDescription, 'SeriesViewerManipulationEventArgsDescription', Description.$);
    return SeriesViewerManipulationEventArgsDescription;
})();
