// igniteui-react-core/esm2015/lib/ShapeSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ShapeSeriesBaseDescription = /*@__PURE__*/ (() => {
    class ShapeSeriesBaseDescription extends SeriesDescription {
        get_type() {
            return "ShapeSeriesBase";
        }
        constructor() {
            super();
            this.e7 = false;
            this.e6 = false;
            this.fk = null;
            this.e3 = null;
            this.e5 = false;
            this.fh = null;
            this.fm = null;
            this.fe = 0;
            this.fd = 0;
            this.fn = null;
            this.fl = null;
            this.fp = null;
            this.fq = null;
            this.fb = 0;
            this.fj = null;
            this.fi = null;
            this.fo = null;
        }
        get isCustomShapeStyleAllowed() {
            return this.e7;
        }
        set isCustomShapeStyleAllowed(a) {
            this.e7 = a;
            this.g("IsCustomShapeStyleAllowed");
        }
        get isCustomShapeMarkerStyleAllowed() {
            return this.e6;
        }
        set isCustomShapeMarkerStyleAllowed(a) {
            this.e6 = a;
            this.g("IsCustomShapeMarkerStyleAllowed");
        }
        get fillMemberPath() {
            return this.fk;
        }
        set fillMemberPath(a) {
            this.fk = a;
            this.g("FillMemberPath");
        }
        get fillScale() {
            return this.e3;
        }
        set fillScale(a) {
            this.e3 = a;
            this.g("FillScale");
        }
        get fillScaleUseGlobalValues() {
            return this.e5;
        }
        set fillScaleUseGlobalValues(a) {
            this.e5 = a;
            this.g("FillScaleUseGlobalValues");
        }
        get actualItemSearchMode() {
            return this.fh;
        }
        set actualItemSearchMode(a) {
            this.fh = a;
            this.g("ActualItemSearchMode");
        }
        get itemSearchMode() {
            return this.fm;
        }
        set itemSearchMode(a) {
            this.fm = a;
            this.g("ItemSearchMode");
        }
        get itemSearchThreshold() {
            return this.fe;
        }
        set itemSearchThreshold(a) {
            this.fe = a;
            this.g("ItemSearchThreshold");
        }
        get itemSearchPointsThreshold() {
            return this.fd;
        }
        set itemSearchPointsThreshold(a) {
            this.fd = a;
            this.g("ItemSearchPointsThreshold");
        }
        get shapeMemberPath() {
            return this.fn;
        }
        set shapeMemberPath(a) {
            this.fn = a;
            this.g("ShapeMemberPath");
        }
        get highlightedShapeMemberPath() {
            return this.fl;
        }
        set highlightedShapeMemberPath(a) {
            this.fl = a;
            this.g("HighlightedShapeMemberPath");
        }
        get xAxisRef() {
            return this.fp;
        }
        set xAxisRef(a) {
            this.fp = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.fq;
        }
        set yAxisRef(a) {
            this.fq = a;
            this.g("YAxisRef");
        }
        get shapeFilterResolution() {
            return this.fb;
        }
        set shapeFilterResolution(a) {
            this.fb = a;
            this.g("ShapeFilterResolution");
        }
        get assigningShapeStyleRef() {
            return this.fj;
        }
        set assigningShapeStyleRef(a) {
            this.fj = a;
            this.g("AssigningShapeStyleRef");
        }
        get assigningShapeMarkerStyleRef() {
            return this.fi;
        }
        set assigningShapeMarkerStyleRef(a) {
            this.fi = a;
            this.g("AssigningShapeMarkerStyleRef");
        }
        get styleShapeRef() {
            return this.fo;
        }
        set styleShapeRef(a) {
            this.fo = a;
            this.g("StyleShapeRef");
        }
    }
    ShapeSeriesBaseDescription.$t = /*@__PURE__*/ markType(ShapeSeriesBaseDescription, 'ShapeSeriesBaseDescription', SeriesDescription.$);
    return ShapeSeriesBaseDescription;
})();
