// igniteui-react-core/esm2015/lib/ShapeFilterRecordEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ShapeFilterRecordEventArgsDescription = /*@__PURE__*/ (() => {
    class ShapeFilterRecordEventArgsDescription extends Description {
        get_type() {
            return "ShapeFilterRecordEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = false;
        }
        get record() {
            return this.h;
        }
        set record(a) {
            this.h = a;
            this.g("Record");
        }
        get shouldInclude() {
            return this.j;
        }
        set shouldInclude(a) {
            this.j = a;
            this.g("ShouldInclude");
        }
    }
    ShapeFilterRecordEventArgsDescription.$t = /*@__PURE__*/ markType(ShapeFilterRecordEventArgsDescription, 'ShapeFilterRecordEventArgsDescription', Description.$);
    return ShapeFilterRecordEventArgsDescription;
})();
