// igniteui-react-core/esm2015/lib/ShapefileRecordDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ShapefileRecordDescription = /*@__PURE__*/ (() => {
    class ShapefileRecordDescription extends Description {
        get_type() {
            return "ShapefileRecord";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = null;
        }
        get fieldValuesRef() {
            return this.j;
        }
        set fieldValuesRef(a) {
            this.j = a;
            this.g("FieldValuesRef");
        }
        get bounds() {
            return this.h;
        }
        set bounds(a) {
            this.h = a;
            this.g("Bounds");
        }
    }
    ShapefileRecordDescription.$t = /*@__PURE__*/ markType(ShapefileRecordDescription, 'ShapefileRecordDescription', Description.$);
    return ShapefileRecordDescription;
})();
