// igniteui-react-core/esm2015/lib/ShapeDataSourceDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ShapeDataSourceDescription = /*@__PURE__*/ (() => {
    class ShapeDataSourceDescription extends Description {
        get_type() {
            return "ShapeDataSource";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
            this.l = false;
            this.r = null;
            this.n = null;
            this.q = null;
            this.p = null;
            this.o = null;
        }
        get worldRect() {
            return this.i;
        }
        set worldRect(a) {
            this.i = a;
            this.g("WorldRect");
        }
        get computedWorldRect() {
            return this.h;
        }
        set computedWorldRect(a) {
            this.h = a;
            this.g("ComputedWorldRect");
        }
        get deferImportCompleted() {
            return this.l;
        }
        set deferImportCompleted(a) {
            this.l = a;
            this.g("DeferImportCompleted");
        }
        get shapefileSource() {
            return this.r;
        }
        set shapefileSource(a) {
            this.r = a;
            this.g("ShapefileSource");
        }
        get databaseSource() {
            return this.n;
        }
        set databaseSource(a) {
            this.n = a;
            this.g("DatabaseSource");
        }
        get importPendingRef() {
            return this.q;
        }
        set importPendingRef(a) {
            this.q = a;
            this.g("ImportPendingRef");
        }
        get importCompletedRef() {
            return this.p;
        }
        set importCompletedRef(a) {
            this.p = a;
            this.g("ImportCompletedRef");
        }
        get filterRef() {
            return this.o;
        }
        set filterRef(a) {
            this.o = a;
            this.g("FilterRef");
        }
    }
    ShapeDataSourceDescription.$t = /*@__PURE__*/ markType(ShapeDataSourceDescription, 'ShapeDataSourceDescription', Description.$);
    return ShapeDataSourceDescription;
})();
