// igniteui-react-core/esm2015/lib/ShapeChartUtils.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, String_$type, markType } from "./type";
import { List$1 } from "./List$1";
import { stringReplace, stringContains } from "./string";
/**
 * @hidden
 */
export let ShapeChartUtils = /*@__PURE__*/ (() => {
    class ShapeChartUtils extends Base {
        static a(a) {
            if (a == 13 || a == 3) {
                return 1;
            }
            if (a == 15 || a == 5) {
                return 2;
            }
            if (a == 1) {
                return 3;
            }
            return 0;
        }
        static ad(a) {
            return ShapeChartUtils.t(ShapeChartUtils.o, a);
        }
        static o(a) {
            a = a.toLowerCase();
            a = stringReplace(stringReplace(a, "[", ""), "]", "");
            return Base.equalsStatic(a, "y") || Base.equalsStatic(a, "yvalue") || Base.equalsStatic(a, "ycolumn") || Base.equalsStatic(a, "ypoint") || Base.equalsStatic(a, "ylocation") || Base.equalsStatic(a, "ycoordinate") || Base.equalsStatic(a, "ydimension") || Base.equalsStatic(a, "lat") || Base.equalsStatic(a, "latitude");
        }
        static ac(a) {
            return ShapeChartUtils.t(ShapeChartUtils.n, a);
        }
        static n(a) {
            a = a.toLowerCase();
            a = stringReplace(stringReplace(a, "[", ""), "]", "");
            return Base.equalsStatic(a, "x") || Base.equalsStatic(a, "xvalue") || Base.equalsStatic(a, "xcolumn") || Base.equalsStatic(a, "xpoint") || Base.equalsStatic(a, "xlocation") || Base.equalsStatic(a, "xcoordinate") || Base.equalsStatic(a, "xdimension") || Base.equalsStatic(a, "lon") || Base.equalsStatic(a, "longitude");
        }
        static u(a) {
            return ShapeChartUtils.t(ShapeChartUtils.f, a);
        }
        static f(a) {
            a = a.toLowerCase();
            a = stringReplace(stringReplace(a, "[", ""), "]", "");
            return Base.equalsStatic(a, "r") || Base.equalsStatic(a, "radius") || Base.equalsStatic(a, "radiusvalue") || Base.equalsStatic(a, "rvalue") || Base.equalsStatic(a, "rcolumn") || Base.equalsStatic(a, "rdimension") || Base.equalsStatic(a, "bubble") || Base.equalsStatic(a, "bubblesize") || Base.equalsStatic(a, "size");
        }
        static aa(a) {
            return ShapeChartUtils.t(ShapeChartUtils.l, a);
        }
        static l(a) {
            a = a.toLowerCase();
            a = stringReplace(stringReplace(a, "[", ""), "]", "");
            return Base.equalsStatic(a, "v") || Base.equalsStatic(a, "value") || Base.equalsStatic(a, "vcolumn") || Base.equalsStatic(a, "z") || Base.equalsStatic(a, "zvalue") || Base.equalsStatic(a, "zcolumn") || Base.equalsStatic(a, "zdimension") || Base.equalsStatic(a, "zcoordinate") || Base.equalsStatic(a, "magnitude") || Base.equalsStatic(a, "elevation") || Base.equalsStatic(a, "depth") || Base.equalsStatic(a, "capacity") || Base.equalsStatic(a, "distance");
        }
        static p(a) {
            return ShapeChartUtils.t(ShapeChartUtils.b, a);
        }
        static b(a) {
            a = a.toLowerCase();
            a = stringReplace(stringReplace(a, "[", ""), "]", "");
            return Base.equalsStatic(a, "f") || Base.equalsStatic(a, "fill") || Base.equalsStatic(a, "fillvalue") || Base.equalsStatic(a, "fvalue") || Base.equalsStatic(a, "fcolumn") || Base.equalsStatic(a, "color") || Base.equalsStatic(a, "colorvalue") || Base.equalsStatic(a, "cvalue") || Base.equalsStatic(a, "ccolumn");
        }
        static s(a) {
            return ShapeChartUtils.t(ShapeChartUtils.e, a);
        }
        static e(a) {
            a = a.toLowerCase();
            a = stringReplace(stringReplace(a, "[", ""), "]", "");
            return Base.equalsStatic(a, "l") || Base.equalsStatic(a, "label") || Base.equalsStatic(a, "lvalue") || Base.equalsStatic(a, "lcolumn") || Base.equalsStatic(a, "n") || Base.equalsStatic(a, "name") || Base.equalsStatic(a, "nvalue") || Base.equalsStatic(a, "ncolumn");
        }
        static r(a) {
            return ShapeChartUtils.t(ShapeChartUtils.d, a);
        }
        static d(a) {
            a = a.toLowerCase();
            a = stringReplace(stringReplace(a, "[", ""), "]", "");
            return Base.equalsStatic(a, "g") || Base.equalsStatic(a, "groupping") || Base.equalsStatic(a, "group") || Base.equalsStatic(a, "gvalue") || Base.equalsStatic(a, "gcolumn") || Base.equalsStatic(a, "region") || Base.equalsStatic(a, "type");
        }
        static y(a) {
            return ShapeChartUtils.t(ShapeChartUtils.j, a);
        }
        static j(a) {
            a = a.toLowerCase();
            a = stringReplace(stringReplace(a, "[", ""), "]", "");
            return Base.equalsStatic(a, "points") || Base.equalsStatic(a, "shapes") || Base.equalsStatic(a, "locations") || Base.equalsStatic(a, "positions") || Base.equalsStatic(a, "coordinates") || Base.equalsStatic(a, "vectors");
        }
        static v(a) {
            return ShapeChartUtils.t(ShapeChartUtils.g, a);
        }
        static g(a) {
            a = a.toLowerCase();
            return stringContains(a, "marginalprofit");
        }
        static w(a) {
            return ShapeChartUtils.t(ShapeChartUtils.h, a);
        }
        static h(a) {
            a = a.toLowerCase();
            return stringContains(a, "revenue") || stringContains(a, "totalsales");
        }
        static q(a) {
            return ShapeChartUtils.t(ShapeChartUtils.c, a);
        }
        static c(a) {
            a = a.toLowerCase();
            return stringContains(a, "fixedcost") || stringContains(a, "fixedexpense");
        }
        static ab(a) {
            return ShapeChartUtils.t(ShapeChartUtils.m, a);
        }
        static m(a) {
            a = a.toLowerCase();
            return stringContains(a, "variablecost") || stringContains(a, "variableexpense");
        }
        static z(a) {
            return ShapeChartUtils.t(ShapeChartUtils.k, a);
        }
        static k(a) {
            a = a.toLowerCase();
            return stringContains(a, "totalcost") || stringContains(a, "totalexpense");
        }
        static x(a) {
            return ShapeChartUtils.t(ShapeChartUtils.i, a);
        }
        static i(a) {
            a = a.toLowerCase();
            return stringContains(a, "units") || stringContains(a, "volume") || stringContains(a, "quantity");
        }
        static t(a, b) {
            let c = new List$1(String_$type, 0);
            for (let d = 0; d < b.count; d++) {
                if (a(b.item(d))) {
                    c.add(b.item(d));
                    b.removeAt(d);
                    d--;
                }
            }
            return c;
        }
    }
    ShapeChartUtils.$t = /*@__PURE__*/ markType(ShapeChartUtils, 'ShapeChartUtils');
    return ShapeChartUtils;
})();
