// igniteui-react-core/esm2015/lib/SizeScaleDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SizeScaleDescription = /*@__PURE__*/ (() => {
    class SizeScaleDescription extends Description {
        get_type() {
            return "SizeScale";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = 0;
            this.j = 0;
            this.m = 0;
            this.l = 0;
            this.h = false;
            this.r = 0;
            this.t = null;
        }
        get globalMinimum() {
            return this.k;
        }
        set globalMinimum(a) {
            this.k = a;
            this.g("GlobalMinimum");
        }
        get globalMaximum() {
            return this.j;
        }
        set globalMaximum(a) {
            this.j = a;
            this.g("GlobalMaximum");
        }
        get minimumValue() {
            return this.m;
        }
        set minimumValue(a) {
            this.m = a;
            this.g("MinimumValue");
        }
        get maximumValue() {
            return this.l;
        }
        set maximumValue(a) {
            this.l = a;
            this.g("MaximumValue");
        }
        get isLogarithmic() {
            return this.h;
        }
        set isLogarithmic(a) {
            this.h = a;
            this.g("IsLogarithmic");
        }
        get logarithmBase() {
            return this.r;
        }
        set logarithmBase(a) {
            this.r = a;
            this.g("LogarithmBase");
        }
        get propertyUpdatedRef() {
            return this.t;
        }
        set propertyUpdatedRef(a) {
            this.t = a;
            this.g("PropertyUpdatedRef");
        }
    }
    SizeScaleDescription.$t = /*@__PURE__*/ markType(SizeScaleDescription, 'SizeScaleDescription', Description.$);
    return SizeScaleDescription;
})();
