// igniteui-react-core/esm2015/lib/SliceClickEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SliceClickEventArgsDescription = /*@__PURE__*/ (() => {
    class SliceClickEventArgsDescription extends Description {
        get_type() {
            return "SliceClickEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = false;
            this.h = false;
            this.l = null;
        }
        get isSelected() {
            return this.i;
        }
        set isSelected(a) {
            this.i = a;
            this.g("IsSelected");
        }
        get isExploded() {
            return this.h;
        }
        set isExploded(a) {
            this.h = a;
            this.g("IsExploded");
        }
        get originalEventRef() {
            return this.l;
        }
        set originalEventRef(a) {
            this.l = a;
            this.g("OriginalEventRef");
        }
    }
    SliceClickEventArgsDescription.$t = /*@__PURE__*/ markType(SliceClickEventArgsDescription, 'SliceClickEventArgsDescription', Description.$);
    return SliceClickEventArgsDescription;
})();
