// igniteui-react-core/esm2015/lib/SliceEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SliceEventArgsDescription = /*@__PURE__*/ (() => {
    class SliceEventArgsDescription extends Description {
        get_type() {
            return "SliceEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = false;
            this.j = false;
            this.h = null;
            this.n = null;
        }
        get isSelected() {
            return this.k;
        }
        set isSelected(a) {
            this.k = a;
            this.g("IsSelected");
        }
        get isExploded() {
            return this.j;
        }
        set isExploded(a) {
            this.j = a;
            this.g("IsExploded");
        }
        get position() {
            return this.h;
        }
        set position(a) {
            this.h = a;
            this.g("Position");
        }
        get originalEventRef() {
            return this.n;
        }
        set originalEventRef(a) {
            this.n = a;
            this.g("OriginalEventRef");
        }
    }
    SliceEventArgsDescription.$t = /*@__PURE__*/ markType(SliceEventArgsDescription, 'SliceEventArgsDescription', Description.$);
    return SliceEventArgsDescription;
})();
