// igniteui-react-core/esm2015/lib/SRProvider.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DependencyObject } from "./DependencyObject";
import { DVSR } from "./DVSR";
import { markType } from "./type";
/**
 * @hidden
 */
export let SRProvider = /*@__PURE__*/ (() => {
    class SRProvider extends DependencyObject {
        get l() {
            return this.k;
        }
        set l(a) {
            this.k = a;
            this.j.container = this.k;
        }
        constructor(a) {
            super();
            this.j = null;
            this.k = null;
            super.toString();
            this.k = a;
            this.j = new DVSR(a);
        }
        cb(a) {
            this.j.setBundleId(a);
        }
        get b3() {
            return this.j.getString("OPD_DefaultInteraction");
        }
        get b4() {
            return this.j.getString("OPD_ScaleToFit");
        }
        get b6() {
            return this.j.getString("OPD_ZoomTo100");
        }
        get b5() {
            return this.j.getString("OPD_ScaleToFit_SeriesViewer");
        }
        get b7() {
            return this.j.getString("PieChart_Others");
        }
        get bz() {
            return this.j.getString("FinancialSeries_Open");
        }
        get bw() {
            return this.j.getString("FinancialSeries_Close");
        }
        get by() {
            return this.j.getString("FinancialSeries_Low");
        }
        get bx() {
            return this.j.getString("FinancialSeries_High");
        }
        get b0() {
            return this.j.getString("FinancialSeries_Volume");
        }
        get m() {
            return this.j.getString("BubbleSeries_Radius");
        }
        get r() {
            return this.j.getString("DataProviderNotInitialized");
        }
        get b2() {
            return this.j.getString("Object_Sealed");
        }
        get b8() {
            return this.j.getString("RangeModificationsNotSupported");
        }
        get s() {
            return this.j.getString("Default_Series_Title");
        }
        get n() {
            return this.j.getString("DataChart_InteractivityNotLoaded");
        }
        get p() {
            return this.j.getString("DataChart_TrendLinesNotLoaded");
        }
        get o() {
            return this.j.getString("DataChart_NumberAbbreviatorNotLoaded");
        }
        get q() {
            return this.j.getString("DataChart_VisualDataNotLoaded");
        }
        get ca() {
            return this.j.getString("TRIAL_VERSION");
        }
        get b9() {
            return this.j.getString("ScatterSeries_Value");
        }
        b1(a) {
            return this.j.getString(a);
        }
        get bq() {
            return this.j.getString("FinancialChart_RangeSelector_OneMonth");
        }
        get bt() {
            return this.j.getString("FinancialChart_RangeSelector_ThreeMonths");
        }
        get bs() {
            return this.j.getString("FinancialChart_RangeSelector_SixMonths");
        }
        get bv() {
            return this.j.getString("FinancialChart_RangeSelector_YearToDate");
        }
        get br() {
            return this.j.getString("FinancialChart_RangeSelector_OneYear");
        }
        get bo() {
            return this.j.getString("FinancialChart_RangeSelector_All");
        }
        get bp() {
            return this.j.getString("FinancialChart_RangeSelector_From");
        }
        get bu() {
            return this.j.getString("FinancialChart_RangeSelector_To");
        }
        get t() {
            return this.j.getString("FinancialChart_IndicatorMenu_Header");
        }
        get v() {
            return this.j.getString("FinancialChart_IndicatorMenu_OverlaysCategoryHeader");
        }
        get w() {
            return this.j.getString("FinancialChart_IndicatorMenu_TrendlinesCategoryHeader");
        }
        get x() {
            return this.j.getString("FinancialChart_IndicatorMenu_VolumeCategoryHeader");
        }
        get u() {
            return this.j.getString("FinancialChart_IndicatorMenu_IndicatorsCategoryHeader");
        }
        get a6() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Overlay_BollingerBands");
        }
        get a7() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Overlay_PriceChannel");
        }
        get y() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_AbsoluteVolumeOscillator");
        }
        get z() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_AccumulationDistribution");
        }
        get aa() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_AverageDirectionalIndex");
        }
        get ab() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_AverageTrueRange");
        }
        get ac() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_BollingerBandWidth");
        }
        get ad() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_ChaikinOscillator");
        }
        get ae() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_ChaikinVolatility");
        }
        get af() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_CommodityChannelIndex");
        }
        get ag() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_DetrendedPriceOscillator");
        }
        get ah() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_EaseOfMovement");
        }
        get ai() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_FastStochasticOscillator");
        }
        get aj() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_ForceIndex");
        }
        get ak() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_FullStochasticOscillator");
        }
        get al() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_MarketFacilitationIndex");
        }
        get am() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_MassIndex");
        }
        get an() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_MedianPrice");
        }
        get ao() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_MoneyFlowIndex");
        }
        get ap() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_MovingAverageConvergenceDivergence");
        }
        get aq() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_NegativeVolumeIndex");
        }
        get ar() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_OnBalanceVolume");
        }
        get as() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_PercentagePriceOscillator");
        }
        get at() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_PercentageVolumeOscillator");
        }
        get au() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_PositiveVolumeIndex");
        }
        get av() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_PriceVolumeTrend");
        }
        get aw() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_RateOfChangeAndMomentum");
        }
        get ax() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_RelativeStrengthIndex");
        }
        get ay() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_SlowStochasticOscillator");
        }
        get az() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_StandardDeviation");
        }
        get a0() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_StochRSI");
        }
        get a1() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_TRIX");
        }
        get a2() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_TypicalPrice");
        }
        get a3() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_UltimateOscillator");
        }
        get a4() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_WeightedClose");
        }
        get a5() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Indicator_WilliamsPercentR");
        }
        get a8() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_CubicFit");
        }
        get a9() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_CumulativeAverage");
        }
        get ba() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_ExponentialAverage");
        }
        get bb() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_ExponentialFit");
        }
        get bc() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_LinearFit");
        }
        get bd() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_LogarithmicFit");
        }
        get be() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_ModifiedAverage");
        }
        get bf() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_PowerLawFit");
        }
        get bg() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_QuadraticFit");
        }
        get bh() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_QuarticFit");
        }
        get bi() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_QuinticFit");
        }
        get bj() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_SimpleAverage");
        }
        get bk() {
            return this.j.getString("FinancialChart_IndicatorsMenu_TrendLine_WeightedAverage");
        }
        get bl() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Volume_Area");
        }
        get bm() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Volume_Column");
        }
        get bn() {
            return this.j.getString("FinancialChart_IndicatorsMenu_Volume_Line");
        }
    }
    SRProvider.$t = /*@__PURE__*/ markType(SRProvider, 'SRProvider', DependencyObject.$);
    return SRProvider;
})();
