// igniteui-react-core/esm2015/lib/StackedSeriesCreatedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let StackedSeriesCreatedEventArgsDescription = /*@__PURE__*/ (() => {
    class StackedSeriesCreatedEventArgsDescription extends Description {
        get_type() {
            return "StackedSeriesCreatedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.r = null;
            this.t = null;
            this.s = null;
            this.u = null;
            this.aa = null;
            this.h = null;
            this.m = 0;
            this.ab = null;
            this.p = 0;
            this.ac = null;
            this.w = null;
            this.x = null;
            this.j = null;
            this.y = null;
            this.z = null;
            this.l = 0;
            this.v = null;
        }
        get brush() {
            return this.r;
        }
        set brush(a) {
            this.r = a;
            this.g("Brush");
        }
        get legendItemTemplateRef() {
            return this.t;
        }
        set legendItemTemplateRef(a) {
            this.t = a;
            this.g("LegendItemTemplateRef");
        }
        get legendItemBadgeTemplateRef() {
            return this.s;
        }
        set legendItemBadgeTemplateRef(a) {
            this.s = a;
            this.g("LegendItemBadgeTemplateRef");
        }
        get legendItemVisibility() {
            return this.u;
        }
        set legendItemVisibility(a) {
            this.u = a;
            this.g("LegendItemVisibility");
        }
        get outline() {
            return this.aa;
        }
        set outline(a) {
            this.aa = a;
            this.g("Outline");
        }
        get dashArray() {
            return this.h;
        }
        set dashArray(a) {
            this.h = a;
            this.g("DashArray");
        }
        get thickness() {
            return this.m;
        }
        set thickness(a) {
            this.m = a;
            this.g("Thickness");
        }
        get titleRef() {
            return this.ab;
        }
        set titleRef(a) {
            this.ab = a;
            this.g("TitleRef");
        }
        get transitionDuration() {
            return this.p;
        }
        set transitionDuration(a) {
            this.p = a;
            this.g("TransitionDuration");
        }
        get transitionEasingFunctionRef() {
            return this.ac;
        }
        set transitionEasingFunctionRef(a) {
            this.ac = a;
            this.g("TransitionEasingFunctionRef");
        }
        get markerBrush() {
            return this.w;
        }
        set markerBrush(a) {
            this.w = a;
            this.g("MarkerBrush");
        }
        get markerOutline() {
            return this.x;
        }
        set markerOutline(a) {
            this.x = a;
            this.g("MarkerOutline");
        }
        get markerStyle() {
            return this.j;
        }
        set markerStyle(a) {
            this.j = a;
            this.g("MarkerStyle");
        }
        get markerTemplateRef() {
            return this.y;
        }
        set markerTemplateRef(a) {
            this.y = a;
            this.g("MarkerTemplateRef");
        }
        get markerType() {
            return this.z;
        }
        set markerType(a) {
            this.z = a;
            this.g("MarkerType");
        }
        get markerThickness() {
            return this.l;
        }
        set markerThickness(a) {
            this.l = a;
            this.g("MarkerThickness");
        }
        get lineCap() {
            return this.v;
        }
        set lineCap(a) {
            this.v = a;
            this.g("LineCap");
        }
    }
    StackedSeriesCreatedEventArgsDescription.$t = /*@__PURE__*/ markType(StackedSeriesCreatedEventArgsDescription, 'StackedSeriesCreatedEventArgsDescription', Description.$);
    return StackedSeriesCreatedEventArgsDescription;
})();
