// igniteui-react-core/esm2015/lib/StackedColumnSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalStackedSeriesBaseDescription } from "./HorizontalStackedSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let StackedColumnSeriesDescription = /*@__PURE__*/ (() => {
    class StackedColumnSeriesDescription extends HorizontalStackedSeriesBaseDescription {
        get_type() {
            return "StackedColumnSeries";
        }
        constructor() {
            super();
            this.gj = 0;
            this.gk = 0;
        }
        get radiusX() {
            return this.gj;
        }
        set radiusX(a) {
            this.gj = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.gk;
        }
        set radiusY(a) {
            this.gk = a;
            this.g("RadiusY");
        }
    }
    StackedColumnSeriesDescription.$t = /*@__PURE__*/ markType(StackedColumnSeriesDescription, 'StackedColumnSeriesDescription', HorizontalStackedSeriesBaseDescription.$);
    return StackedColumnSeriesDescription;
})();
