// igniteui-react-core/esm2015/lib/StockDetails.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StockInfo } from "./StockInfo";
import { markType } from "./type";
import { round10N } from "./numberExtended";
/**
 * @hidden
 */
export let StockDetails = /*@__PURE__*/ (() => {
    class StockDetails extends StockInfo {
        constructor() {
            super(...arguments);
            this._annualLow = 0;
            this._annualHigh = 0;
            this._marketPrice = 0;
            this._marketShares = 0;
            this._volatility = 0;
        }
        get annualLow() {
            return this._annualLow;
        }
        set annualLow(a) {
            this._annualLow = a;
        }
        get annualHigh() {
            return this._annualHigh;
        }
        set annualHigh(a) {
            this._annualHigh = a;
        }
        get marketPrice() {
            return this._marketPrice;
        }
        set marketPrice(a) {
            this._marketPrice = a;
        }
        get marketShares() {
            return this._marketShares;
        }
        set marketShares(a) {
            this._marketShares = a;
        }
        get marketCap() {
            return round10N(this.marketPrice * this.marketShares / 1000, 1);
        }
        get volatility() {
            return this._volatility;
        }
        set volatility(a) {
            this._volatility = a;
        }
    }
    StockDetails.$t = /*@__PURE__*/ markType(StockDetails, 'StockDetails', StockInfo.$);
    return StockDetails;
})();
