// igniteui-react-core/esm2015/lib/StockPricePoint.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StockInfo } from "./StockInfo";
import { markType } from "./type";
/**
 * @hidden
 */
export let StockPricePoint = /*@__PURE__*/ (() => {
    class StockPricePoint extends StockInfo {
        constructor() {
            super();
            this._open = 0;
            this._high = 0;
            this._low = 0;
            this._close = 0;
            this._volume = 0;
            this._date = new Date();
        }
        get open() {
            return this._open;
        }
        set open(a) {
            this._open = a;
        }
        get high() {
            return this._high;
        }
        set high(a) {
            this._high = a;
        }
        get low() {
            return this._low;
        }
        set low(a) {
            this._low = a;
        }
        get close() {
            return this._close;
        }
        set close(a) {
            this._close = a;
        }
        get volume() {
            return this._volume;
        }
        set volume(a) {
            this._volume = a;
        }
        get date() {
            return this._date;
        }
        set date(a) {
            this._date = a;
        }
    }
    StockPricePoint.$t = /*@__PURE__*/ markType(StockPricePoint, 'StockPricePoint', StockInfo.$);
    return StockPricePoint;
})();
