// igniteui-react-core/esm2015/lib/StyleShapeEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let StyleShapeEventArgsDescription = /*@__PURE__*/ (() => {
    class StyleShapeEventArgsDescription extends Description {
        get_type() {
            return "StyleShapeEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.m = null;
            this.n = null;
            this.i = 0;
            this.h = 0;
        }
        get itemRef() {
            return this.l;
        }
        set itemRef(a) {
            this.l = a;
            this.g("ItemRef");
        }
        get shapeFill() {
            return this.m;
        }
        set shapeFill(a) {
            this.m = a;
            this.g("ShapeFill");
        }
        get shapeStroke() {
            return this.n;
        }
        set shapeStroke(a) {
            this.n = a;
            this.g("ShapeStroke");
        }
        get shapeStrokeThickness() {
            return this.i;
        }
        set shapeStrokeThickness(a) {
            this.i = a;
            this.g("ShapeStrokeThickness");
        }
        get shapeOpacity() {
            return this.h;
        }
        set shapeOpacity(a) {
            this.h = a;
            this.g("ShapeOpacity");
        }
    }
    StyleShapeEventArgsDescription.$t = /*@__PURE__*/ markType(StyleShapeEventArgsDescription, 'StyleShapeEventArgsDescription', Description.$);
    return StyleShapeEventArgsDescription;
})();
