// igniteui-react-core/esm2015/lib/SuffixDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { InputGroupItemDescription } from "./InputGroupItemDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let SuffixDescription = /*@__PURE__*/ (() => {
    class SuffixDescription extends InputGroupItemDescription {
        get_type() {
            return "Suffix";
        }
        constructor() {
            super();
            this.p = null;
            this.o = null;
            this.l = false;
            this.k = false;
        }
        get id() {
            return this.p;
        }
        set id(a) {
            this.p = a;
            this.g("Id");
        }
        get ariaLabel() {
            return this.o;
        }
        set ariaLabel(a) {
            this.o = a;
            this.g("AriaLabel");
        }
        get isHover() {
            return this.l;
        }
        set isHover(a) {
            this.l = a;
            this.g("IsHover");
        }
        get disabled() {
            return this.k;
        }
        set disabled(a) {
            this.k = a;
            this.g("Disabled");
        }
    }
    SuffixDescription.$t = /*@__PURE__*/ markType(SuffixDescription, 'SuffixDescription', InputGroupItemDescription.$);
    return SuffixDescription;
})();
