// igniteui-react-core/esm2015/lib/SummaryRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DefinitionBaseDescription } from "./DefinitionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let SummaryRowDescription = /*@__PURE__*/ (() => {
    class SummaryRowDescription extends DefinitionBaseDescription {
        get_type() {
            return "SummaryRow";
        }
        constructor() {
            super();
            this.c0 = null;
            this.cy = null;
            this.c4 = null;
            this.cz = null;
            this.c1 = null;
            this.co = 0;
            this.c2 = null;
            this.c3 = null;
            this.cr = 0;
            this.ct = 0;
            this.cs = 0;
            this.cq = 0;
        }
        get selectedBackground() {
            return this.c0;
        }
        set selectedBackground(a) {
            this.c0 = a;
            this.g("SelectedBackground");
        }
        get actualSelectedBackground() {
            return this.cy;
        }
        set actualSelectedBackground(a) {
            this.cy = a;
            this.g("ActualSelectedBackground");
        }
        get summaryLabelTextColor() {
            return this.c4;
        }
        set summaryLabelTextColor(a) {
            this.c4 = a;
            this.g("SummaryLabelTextColor");
        }
        get actualSummaryLabelTextColor() {
            return this.cz;
        }
        set actualSummaryLabelTextColor(a) {
            this.cz = a;
            this.g("ActualSummaryLabelTextColor");
        }
        get summaryLabelFontFamily() {
            return this.c1;
        }
        set summaryLabelFontFamily(a) {
            this.c1 = a;
            this.g("SummaryLabelFontFamily");
        }
        get summaryLabelFontSize() {
            return this.co;
        }
        set summaryLabelFontSize(a) {
            this.co = a;
            this.g("SummaryLabelFontSize");
        }
        get summaryLabelFontStyle() {
            return this.c2;
        }
        set summaryLabelFontStyle(a) {
            this.c2 = a;
            this.g("SummaryLabelFontStyle");
        }
        get summaryLabelFontWeight() {
            return this.c3;
        }
        set summaryLabelFontWeight(a) {
            this.c3 = a;
            this.g("SummaryLabelFontWeight");
        }
        get paddingLeft() {
            return this.cr;
        }
        set paddingLeft(a) {
            this.cr = a;
            this.g("PaddingLeft");
        }
        get paddingTop() {
            return this.ct;
        }
        set paddingTop(a) {
            this.ct = a;
            this.g("PaddingTop");
        }
        get paddingRight() {
            return this.cs;
        }
        set paddingRight(a) {
            this.cs = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.cq;
        }
        set paddingBottom(a) {
            this.cq = a;
            this.g("PaddingBottom");
        }
    }
    SummaryRowDescription.$t = /*@__PURE__*/ markType(SummaryRowDescription, 'SummaryRowDescription', DefinitionBaseDescription.$);
    return SummaryRowDescription;
})();
