// igniteui-react-core/esm2015/lib/SummaryCellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CellInfoDescription } from "./CellInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let SummaryCellInfoDescription = /*@__PURE__*/ (() => {
    class SummaryCellInfoDescription extends CellInfoDescription {
        get_type() {
            return "SummaryCellInfo";
        }
        constructor() {
            super();
            this.fp = null;
            this.fq = null;
            this.fu = null;
            this.fr = null;
            this.fm = 0;
            this.fs = null;
            this.ft = null;
        }
        get resolvedSummaryLabel() {
            return this.fp;
        }
        set resolvedSummaryLabel(a) {
            this.fp = a;
            this.g("ResolvedSummaryLabel");
        }
        get resolvedSummaryValue() {
            return this.fq;
        }
        set resolvedSummaryValue(a) {
            this.fq = a;
            this.g("ResolvedSummaryValue");
        }
        get summaryLabelTextColor() {
            return this.fu;
        }
        set summaryLabelTextColor(a) {
            this.fu = a;
            this.g("SummaryLabelTextColor");
        }
        get summaryLabelFontFamily() {
            return this.fr;
        }
        set summaryLabelFontFamily(a) {
            this.fr = a;
            this.g("SummaryLabelFontFamily");
        }
        get summaryLabelFontSize() {
            return this.fm;
        }
        set summaryLabelFontSize(a) {
            this.fm = a;
            this.g("SummaryLabelFontSize");
        }
        get summaryLabelFontStyle() {
            return this.fs;
        }
        set summaryLabelFontStyle(a) {
            this.fs = a;
            this.g("SummaryLabelFontStyle");
        }
        get summaryLabelFontWeight() {
            return this.ft;
        }
        set summaryLabelFontWeight(a) {
            this.ft = a;
            this.g("SummaryLabelFontWeight");
        }
    }
    SummaryCellInfoDescription.$t = /*@__PURE__*/ markType(SummaryCellInfoDescription, 'SummaryCellInfoDescription', CellInfoDescription.$);
    SummaryCellInfoDescription.__marshalByValue1 = true;
    SummaryCellInfoDescription.__marshalByValueAlias1 = "SummaryCellModel";
    return SummaryCellInfoDescription;
})();
