// igniteui-react-core/esm2015/lib/SummaryDataDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SummaryDataDescription = /*@__PURE__*/ (() => {
    class SummaryDataDescription extends Description {
        get_type() {
            return "SummaryData";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.l = null;
            this.k = null;
            this.h = null;
            this.i = null;
        }
        get summaryName() {
            return this.j;
        }
        set summaryName(a) {
            this.j = a;
            this.g("SummaryName");
        }
        get summaryValueRef() {
            return this.l;
        }
        set summaryValueRef(a) {
            this.l = a;
            this.g("SummaryValueRef");
        }
        get summaryOperand() {
            return this.k;
        }
        set summaryOperand(a) {
            this.k = a;
            this.g("SummaryOperand");
        }
        get formattedText() {
            return this.h;
        }
        set formattedText(a) {
            this.h = a;
            this.g("FormattedText");
        }
        get formattedValue() {
            return this.i;
        }
        set formattedValue(a) {
            this.i = a;
            this.g("FormattedValue");
        }
    }
    SummaryDataDescription.$t = /*@__PURE__*/ markType(SummaryDataDescription, 'SummaryDataDescription', Description.$);
    return SummaryDataDescription;
})();
