// igniteui-react-core/esm2015/lib/TemplateColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DataGridColumnDescription } from "./DataGridColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let TemplateColumnDescription = /*@__PURE__*/ (() => {
    class TemplateColumnDescription extends DataGridColumnDescription {
        get templateRef() {
            return this.f4;
        }
        set templateRef(a) {
            this.f4 = a;
            this.g("TemplateRef");
        }
        get_type() {
            return "TemplateColumn";
        }
        constructor() {
            super();
            this.f4 = null;
            this.f3 = null;
        }
        get cellUpdatingRef() {
            return this.f3;
        }
        set cellUpdatingRef(a) {
            this.f3 = a;
            this.g("CellUpdatingRef");
        }
    }
    TemplateColumnDescription.$t = /*@__PURE__*/ markType(TemplateColumnDescription, 'TemplateColumnDescription', DataGridColumnDescription.$);
    TemplateColumnDescription.__canMarshalByValue1 = true;
    TemplateColumnDescription.__marshalByValueAlias1 = "TemplateColumn";
    return TemplateColumnDescription;
})();
