// igniteui-react-core/esm2015/lib/TextColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DataGridColumnDescription } from "./DataGridColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let TextColumnDescription = /*@__PURE__*/ (() => {
    class TextColumnDescription extends DataGridColumnDescription {
        get_type() {
            return "TextColumn";
        }
        constructor() {
            super();
            this.f7 = null;
            this.f5 = null;
            this.f1 = null;
            this.f6 = null;
            this.f8 = null;
        }
        get editorType() {
            return this.f7;
        }
        set editorType(a) {
            this.f7 = a;
            this.g("EditorType");
        }
        get editorDataSourceRef() {
            return this.f5;
        }
        set editorDataSourceRef(a) {
            this.f5 = a;
            this.g("EditorDataSourceRef");
        }
        get editorDataSource() {
            return this.f1;
        }
        set editorDataSource(a) {
            this.f1 = a;
            this.g("EditorDataSource");
        }
        get editorTextField() {
            return this.f6;
        }
        set editorTextField(a) {
            this.f6 = a;
            this.g("EditorTextField");
        }
        get editorValueField() {
            return this.f8;
        }
        set editorValueField(a) {
            this.f8 = a;
            this.g("EditorValueField");
        }
    }
    TextColumnDescription.$t = /*@__PURE__*/ markType(TextColumnDescription, 'TextColumnDescription', DataGridColumnDescription.$);
    TextColumnDescription.__canMarshalByValue1 = true;
    TextColumnDescription.__marshalByValueAlias1 = "TextColumn";
    return TextColumnDescription;
})();
