// igniteui-react-core/esm2015/lib/TimeAxisBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategoryAxisBaseDescription } from "./CategoryAxisBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let TimeAxisBaseDescription = /*@__PURE__*/ (() => {
    class TimeAxisBaseDescription extends CategoryAxisBaseDescription {
        get_type() {
            return "TimeAxisBase";
        }
        constructor() {
            super();
            this.ef = null;
            this.d3 = false;
            this.d6 = new Date();
            this.d5 = new Date();
            this.d8 = new Date();
            this.d7 = new Date();
            this.ee = null;
            this.ed = null;
        }
        get dateTimeMemberPath() {
            return this.ef;
        }
        set dateTimeMemberPath(a) {
            this.ef = a;
            this.g("DateTimeMemberPath");
        }
        get isDataPreSorted() {
            return this.d3;
        }
        set isDataPreSorted(a) {
            this.d3 = a;
            this.g("IsDataPreSorted");
        }
        get actualMinimumValue() {
            return this.d6;
        }
        set actualMinimumValue(a) {
            this.d6 = a;
            this.g("ActualMinimumValue");
        }
        get actualMaximumValue() {
            return this.d5;
        }
        set actualMaximumValue(a) {
            this.d5 = a;
            this.g("ActualMaximumValue");
        }
        get minimumValue() {
            return this.d8;
        }
        set minimumValue(a) {
            this.d8 = a;
            this.g("MinimumValue");
        }
        get maximumValue() {
            return this.d7;
        }
        set maximumValue(a) {
            this.d7 = a;
            this.g("MaximumValue");
        }
        get actualMinimumValueChangeRef() {
            return this.ee;
        }
        set actualMinimumValueChangeRef(a) {
            this.ee = a;
            this.g("ActualMinimumValueChangeRef");
        }
        get actualMaximumValueChangeRef() {
            return this.ed;
        }
        set actualMaximumValueChangeRef(a) {
            this.ed = a;
            this.g("ActualMaximumValueChangeRef");
        }
    }
    TimeAxisBaseDescription.$t = /*@__PURE__*/ markType(TimeAxisBaseDescription, 'TimeAxisBaseDescription', CategoryAxisBaseDescription.$);
    return TimeAxisBaseDescription;
})();
