// igniteui-react-core/esm2015/lib/TimeXAxisDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { TimeAxisBaseDescription } from "./TimeAxisBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let TimeXAxisDescription = /*@__PURE__*/ (() => {
    class TimeXAxisDescription extends TimeAxisBaseDescription {
        get_type() {
            return "TimeXAxis";
        }
        constructor() {
            super();
            this.ej = null;
            this.ep = null;
            this.el = null;
            this.ek = null;
        }
        get breaks() {
            return this.ej;
        }
        set breaks(a) {
            this.ej = a;
            this.g("Breaks");
        }
        get labellingMode() {
            return this.ep;
        }
        set labellingMode(a) {
            this.ep = a;
            this.g("LabellingMode");
        }
        get labelFormats() {
            return this.el;
        }
        set labelFormats(a) {
            this.el = a;
            this.g("LabelFormats");
        }
        get intervals() {
            return this.ek;
        }
        set intervals(a) {
            this.ek = a;
            this.g("Intervals");
        }
    }
    TimeXAxisDescription.$t = /*@__PURE__*/ markType(TimeXAxisDescription, 'TimeXAxisDescription', TimeAxisBaseDescription.$);
    return TimeXAxisDescription;
})();
