// igniteui-react-core/esm2015/lib/TimeAxisLabelFormatDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let TimeAxisLabelFormatDescription = /*@__PURE__*/ (() => {
    class TimeAxisLabelFormatDescription extends Description {
        get_type() {
            return "TimeAxisLabelFormat";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = 0;
            this.j = null;
            this.m = null;
            this.l = null;
            this.k = null;
        }
        get range() {
            return this.h;
        }
        set range(a) {
            this.h = a;
            this.g("Range");
        }
        get format() {
            return this.j;
        }
        set format(a) {
            this.j = a;
            this.g("Format");
        }
        get repeatedYearFormat() {
            return this.m;
        }
        set repeatedYearFormat(a) {
            this.m = a;
            this.g("RepeatedYearFormat");
        }
        get repeatedMonthFormat() {
            return this.l;
        }
        set repeatedMonthFormat(a) {
            this.l = a;
            this.g("RepeatedMonthFormat");
        }
        get repeatedDayFormat() {
            return this.k;
        }
        set repeatedDayFormat(a) {
            this.k = a;
            this.g("RepeatedDayFormat");
        }
    }
    TimeAxisLabelFormatDescription.$t = /*@__PURE__*/ markType(TimeAxisLabelFormatDescription, 'TimeAxisLabelFormatDescription', Description.$);
    return TimeAxisLabelFormatDescription;
})();
