// igniteui-react-core/esm2015/lib/ToolActionPopupOpeningEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionPopupOpeningEventArgsDescription = /*@__PURE__*/ (() => {
    class ToolActionPopupOpeningEventArgsDescription extends Description {
        get_type() {
            return "ToolActionPopupOpeningEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = false;
        }
        get sourceAction() {
            return this.h;
        }
        set sourceAction(a) {
            this.h = a;
            this.g("SourceAction");
        }
        get cancel() {
            return this.j;
        }
        set cancel(a) {
            this.j = a;
            this.g("Cancel");
        }
    }
    ToolActionPopupOpeningEventArgsDescription.$t = /*@__PURE__*/ markType(ToolActionPopupOpeningEventArgsDescription, 'ToolActionPopupOpeningEventArgsDescription', Description.$);
    return ToolActionPopupOpeningEventArgsDescription;
})();
