// igniteui-react-core/esm2015/lib/ToolActionComboDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionComboDescription = /*@__PURE__*/ (() => {
    class ToolActionComboDescription extends ToolActionDescription {
        get_type() {
            return "ToolActionCombo";
        }
        constructor() {
            super();
            this.dr = null;
            this.ds = null;
            this.dt = null;
            this.dp = null;
        }
        get dataSourceRef() {
            return this.dr;
        }
        set dataSourceRef(a) {
            this.dr = a;
            this.g("DataSourceRef");
        }
        get displayMemberPath() {
            return this.ds;
        }
        set displayMemberPath(a) {
            this.ds = a;
            this.g("DisplayMemberPath");
        }
        get valueMemberPath() {
            return this.dt;
        }
        set valueMemberPath(a) {
            this.dt = a;
            this.g("ValueMemberPath");
        }
        get selectedValues() {
            return this.dp;
        }
        set selectedValues(a) {
            this.dp = a;
            this.g("SelectedValues");
        }
    }
    ToolActionComboDescription.$t = /*@__PURE__*/ markType(ToolActionComboDescription, 'ToolActionComboDescription', ToolActionDescription.$);
    return ToolActionComboDescription;
})();
