// igniteui-react-core/esm2015/lib/ToolActionCheckboxListInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionInfoDescription } from "./ToolActionInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionCheckboxListInfoDescription = /*@__PURE__*/ (() => {
    class ToolActionCheckboxListInfoDescription extends ToolActionInfoDescription {
        get_type() {
            return "ToolActionCheckboxListInfo";
        }
        constructor() {
            super();
            this.bj = null;
            this.bi = null;
            this.bk = null;
            this.bf = false;
            this.bc = null;
        }
        get dataSourceRef() {
            return this.bj;
        }
        set dataSourceRef(a) {
            this.bj = a;
            this.g("DataSourceRef");
        }
        get dataMemberPath() {
            return this.bi;
        }
        set dataMemberPath(a) {
            this.bi = a;
            this.g("DataMemberPath");
        }
        get selectedMemberPath() {
            return this.bk;
        }
        set selectedMemberPath(a) {
            this.bk = a;
            this.g("SelectedMemberPath");
        }
        get showSelectAll() {
            return this.bf;
        }
        set showSelectAll(a) {
            this.bf = a;
            this.g("ShowSelectAll");
        }
        get primaryKey() {
            return this.bc;
        }
        set primaryKey(a) {
            this.bc = a;
            this.g("PrimaryKey");
        }
    }
    ToolActionCheckboxListInfoDescription.$t = /*@__PURE__*/ markType(ToolActionCheckboxListInfoDescription, 'ToolActionCheckboxListInfoDescription', ToolActionInfoDescription.$);
    ToolActionCheckboxListInfoDescription.__marshalByValue1 = true;
    ToolActionCheckboxListInfoDescription.__marshalByValueAlias1 = "ToolActionCheckboxListInfo";
    return ToolActionCheckboxListInfoDescription;
})();
