// igniteui-react-core/esm2015/lib/ToolPanelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolPanelDescription = /*@__PURE__*/ (() => {
    class ToolPanelDescription extends Description {
        get_type() {
            return "ToolPanel";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = false;
            this.k = false;
            this.be = 0;
            this.b9 = null;
            this.z = 0;
            this.ca = null;
            this.cb = null;
            this.cv = null;
            this.ak = 0;
            this.cw = null;
            this.cx = null;
            this.b8 = null;
            this.cf = null;
            this.bo = null;
            this.b7 = null;
            this.bl = null;
            this.cu = null;
            this.b0 = null;
            this.ch = null;
            this.bq = null;
            this.ct = null;
            this.bz = null;
            this.ci = null;
            this.br = null;
            this.cm = null;
            this.bv = null;
            this.cz = null;
            this.b2 = null;
            this.b1 = null;
            this.cy = null;
            this.cg = null;
            this.bp = null;
            this.ck = null;
            this.bt = null;
            this.cj = null;
            this.bs = null;
            this.c1 = null;
            this.b4 = null;
            this.c0 = null;
            this.b3 = null;
            this.al = 0;
            this.x = 0;
            this.am = 0;
            this.y = 0;
            this.cn = null;
            this.bw = null;
            this.co = null;
            this.bx = null;
            this.ah = 0;
            this.u = 0;
            this.ai = 0;
            this.v = 0;
            this.aj = 0;
            this.w = 0;
            this.ag = 0;
            this.t = 0;
            this.ad = 0;
            this.q = 0;
            this.ae = 0;
            this.r = 0;
            this.af = 0;
            this.s = 0;
            this.ac = 0;
            this.p = 0;
            this.cs = null;
            this.bg = 0;
            this.ab = 0;
            this.o = 0;
            this.aa = 0;
            this.n = 0;
            this.cl = null;
            this.bu = null;
            this.cp = null;
            this.by = null;
            this.c2 = null;
            this.b5 = null;
            this.cc = null;
            this.bm = null;
            this.c3 = null;
            this.b6 = null;
            this.cd = null;
            this.bn = null;
            this.bf = 0;
            this.bd = 0;
            this.cq = null;
            this.ce = null;
            this.cr = null;
        }
        get actions() {
            return this.h;
        }
        set actions(a) {
            this.h = a;
            this.g("Actions");
        }
        get showOnHover() {
            return this.j;
        }
        set showOnHover(a) {
            this.j = a;
            this.g("ShowOnHover");
        }
        get showTooltipOnHover() {
            return this.k;
        }
        set showTooltipOnHover(a) {
            this.k = a;
            this.g("ShowTooltipOnHover");
        }
        get dropdownClickBuffer() {
            return this.be;
        }
        set dropdownClickBuffer(a) {
            this.be = a;
            this.g("DropdownClickBuffer");
        }
        get cellFontFamily() {
            return this.b9;
        }
        set cellFontFamily(a) {
            this.b9 = a;
            this.g("CellFontFamily");
        }
        get cellFontSize() {
            return this.z;
        }
        set cellFontSize(a) {
            this.z = a;
            this.g("CellFontSize");
        }
        get cellFontStyle() {
            return this.ca;
        }
        set cellFontStyle(a) {
            this.ca = a;
            this.g("CellFontStyle");
        }
        get cellFontWeight() {
            return this.cb;
        }
        set cellFontWeight(a) {
            this.cb = a;
            this.g("CellFontWeight");
        }
        get subtitleFontFamily() {
            return this.cv;
        }
        set subtitleFontFamily(a) {
            this.cv = a;
            this.g("SubtitleFontFamily");
        }
        get subtitleFontSize() {
            return this.ak;
        }
        set subtitleFontSize(a) {
            this.ak = a;
            this.g("SubtitleFontSize");
        }
        get subtitleFontStyle() {
            return this.cw;
        }
        set subtitleFontStyle(a) {
            this.cw = a;
            this.g("SubtitleFontStyle");
        }
        get subtitleFontWeight() {
            return this.cx;
        }
        set subtitleFontWeight(a) {
            this.cx = a;
            this.g("SubtitleFontWeight");
        }
        get baseTheme() {
            return this.b8;
        }
        set baseTheme(a) {
            this.b8 = a;
            this.g("BaseTheme");
        }
        get density() {
            return this.cf;
        }
        set density(a) {
            this.cf = a;
            this.g("Density");
        }
        get actualDensity() {
            return this.bo;
        }
        set actualDensity(a) {
            this.bo = a;
            this.g("ActualDensity");
        }
        get backgroundColor() {
            return this.b7;
        }
        set backgroundColor(a) {
            this.b7 = a;
            this.g("BackgroundColor");
        }
        get actualBackgroundColor() {
            return this.bl;
        }
        set actualBackgroundColor(a) {
            this.bl = a;
            this.g("ActualBackgroundColor");
        }
        get submenuBackgroundColor() {
            return this.cu;
        }
        set submenuBackgroundColor(a) {
            this.cu = a;
            this.g("SubmenuBackgroundColor");
        }
        get actualSubmenuBackgroundColor() {
            return this.b0;
        }
        set actualSubmenuBackgroundColor(a) {
            this.b0 = a;
            this.g("ActualSubmenuBackgroundColor");
        }
        get groupHeaderBackgroundColor() {
            return this.ch;
        }
        set groupHeaderBackgroundColor(a) {
            this.ch = a;
            this.g("GroupHeaderBackgroundColor");
        }
        get actualGroupHeaderBackgroundColor() {
            return this.bq;
        }
        set actualGroupHeaderBackgroundColor(a) {
            this.bq = a;
            this.g("ActualGroupHeaderBackgroundColor");
        }
        get separatorBackgroundColor() {
            return this.ct;
        }
        set separatorBackgroundColor(a) {
            this.ct = a;
            this.g("SeparatorBackgroundColor");
        }
        get actualSeparatorBackgroundColor() {
            return this.bz;
        }
        set actualSeparatorBackgroundColor(a) {
            this.bz = a;
            this.g("ActualSeparatorBackgroundColor");
        }
        get groupHeaderSeparatorBackgroundColor() {
            return this.ci;
        }
        set groupHeaderSeparatorBackgroundColor(a) {
            this.ci = a;
            this.g("GroupHeaderSeparatorBackgroundColor");
        }
        get actualGroupHeaderSeparatorBackgroundColor() {
            return this.br;
        }
        set actualGroupHeaderSeparatorBackgroundColor(a) {
            this.br = a;
            this.g("ActualGroupHeaderSeparatorBackgroundColor");
        }
        get hoverBackgroundColor() {
            return this.cm;
        }
        set hoverBackgroundColor(a) {
            this.cm = a;
            this.g("HoverBackgroundColor");
        }
        get actualHoverBackgroundColor() {
            return this.bv;
        }
        set actualHoverBackgroundColor(a) {
            this.bv = a;
            this.g("ActualHoverBackgroundColor");
        }
        get textColor() {
            return this.cz;
        }
        set textColor(a) {
            this.cz = a;
            this.g("TextColor");
        }
        get actualTextColor() {
            return this.b2;
        }
        set actualTextColor(a) {
            this.b2 = a;
            this.g("ActualTextColor");
        }
        get actualSubtitleTextColor() {
            return this.b1;
        }
        set actualSubtitleTextColor(a) {
            this.b1 = a;
            this.g("ActualSubtitleTextColor");
        }
        get subtitleTextColor() {
            return this.cy;
        }
        set subtitleTextColor(a) {
            this.cy = a;
            this.g("SubtitleTextColor");
        }
        get disabledTextColor() {
            return this.cg;
        }
        set disabledTextColor(a) {
            this.cg = a;
            this.g("DisabledTextColor");
        }
        get actualDisabledTextColor() {
            return this.bp;
        }
        set actualDisabledTextColor(a) {
            this.bp = a;
            this.g("ActualDisabledTextColor");
        }
        get groupHeaderTextColor() {
            return this.ck;
        }
        set groupHeaderTextColor(a) {
            this.ck = a;
            this.g("GroupHeaderTextColor");
        }
        get actualGroupHeaderTextColor() {
            return this.bt;
        }
        set actualGroupHeaderTextColor(a) {
            this.bt = a;
            this.g("ActualGroupHeaderTextColor");
        }
        get groupHeaderSubtitleTextColor() {
            return this.cj;
        }
        set groupHeaderSubtitleTextColor(a) {
            this.cj = a;
            this.g("GroupHeaderSubtitleTextColor");
        }
        get actualGroupHeaderSubtitleTextColor() {
            return this.bs;
        }
        set actualGroupHeaderSubtitleTextColor(a) {
            this.bs = a;
            this.g("ActualGroupHeaderSubtitleTextColor");
        }
        get toolTipTextColor() {
            return this.c1;
        }
        set toolTipTextColor(a) {
            this.c1 = a;
            this.g("ToolTipTextColor");
        }
        get actualToolTipTextColor() {
            return this.b4;
        }
        set actualToolTipTextColor(a) {
            this.b4 = a;
            this.g("ActualToolTipTextColor");
        }
        get toolTipBackgroundColor() {
            return this.c0;
        }
        set toolTipBackgroundColor(a) {
            this.c0 = a;
            this.g("ToolTipBackgroundColor");
        }
        get actualToolTipBackgroundColor() {
            return this.b3;
        }
        set actualToolTipBackgroundColor(a) {
            this.b3 = a;
            this.g("ActualToolTipBackgroundColor");
        }
        get toolTipCornerRadius() {
            return this.al;
        }
        set toolTipCornerRadius(a) {
            this.al = a;
            this.g("ToolTipCornerRadius");
        }
        get actualToolTipCornerRadius() {
            return this.x;
        }
        set actualToolTipCornerRadius(a) {
            this.x = a;
            this.g("ActualToolTipCornerRadius");
        }
        get toolTipElevation() {
            return this.am;
        }
        set toolTipElevation(a) {
            this.am = a;
            this.g("ToolTipElevation");
        }
        get actualToolTipElevation() {
            return this.y;
        }
        set actualToolTipElevation(a) {
            this.y = a;
            this.g("ActualToolTipElevation");
        }
        get iconFill() {
            return this.cn;
        }
        set iconFill(a) {
            this.cn = a;
            this.g("IconFill");
        }
        get actualIconFill() {
            return this.bw;
        }
        set actualIconFill(a) {
            this.bw = a;
            this.g("ActualIconFill");
        }
        get iconStroke() {
            return this.co;
        }
        set iconStroke(a) {
            this.co = a;
            this.g("IconStroke");
        }
        get actualIconStroke() {
            return this.bx;
        }
        set actualIconStroke(a) {
            this.bx = a;
            this.g("ActualIconStroke");
        }
        get separatorVerticalPaddingLeft() {
            return this.ah;
        }
        set separatorVerticalPaddingLeft(a) {
            this.ah = a;
            this.g("SeparatorVerticalPaddingLeft");
        }
        get actualSeparatorVerticalPaddingLeft() {
            return this.u;
        }
        set actualSeparatorVerticalPaddingLeft(a) {
            this.u = a;
            this.g("ActualSeparatorVerticalPaddingLeft");
        }
        get separatorVerticalPaddingRight() {
            return this.ai;
        }
        set separatorVerticalPaddingRight(a) {
            this.ai = a;
            this.g("SeparatorVerticalPaddingRight");
        }
        get actualSeparatorVerticalPaddingRight() {
            return this.v;
        }
        set actualSeparatorVerticalPaddingRight(a) {
            this.v = a;
            this.g("ActualSeparatorVerticalPaddingRight");
        }
        get separatorVerticalPaddingTop() {
            return this.aj;
        }
        set separatorVerticalPaddingTop(a) {
            this.aj = a;
            this.g("SeparatorVerticalPaddingTop");
        }
        get actualSeparatorVerticalPaddingTop() {
            return this.w;
        }
        set actualSeparatorVerticalPaddingTop(a) {
            this.w = a;
            this.g("ActualSeparatorVerticalPaddingTop");
        }
        get separatorVerticalPaddingBottom() {
            return this.ag;
        }
        set separatorVerticalPaddingBottom(a) {
            this.ag = a;
            this.g("SeparatorVerticalPaddingBottom");
        }
        get actualSeparatorVerticalPaddingBottom() {
            return this.t;
        }
        set actualSeparatorVerticalPaddingBottom(a) {
            this.t = a;
            this.g("ActualSeparatorVerticalPaddingBottom");
        }
        get separatorHorizontalPaddingLeft() {
            return this.ad;
        }
        set separatorHorizontalPaddingLeft(a) {
            this.ad = a;
            this.g("SeparatorHorizontalPaddingLeft");
        }
        get actualSeparatorHorizontalPaddingLeft() {
            return this.q;
        }
        set actualSeparatorHorizontalPaddingLeft(a) {
            this.q = a;
            this.g("ActualSeparatorHorizontalPaddingLeft");
        }
        get separatorHorizontalPaddingRight() {
            return this.ae;
        }
        set separatorHorizontalPaddingRight(a) {
            this.ae = a;
            this.g("SeparatorHorizontalPaddingRight");
        }
        get actualSeparatorHorizontalPaddingRight() {
            return this.r;
        }
        set actualSeparatorHorizontalPaddingRight(a) {
            this.r = a;
            this.g("ActualSeparatorHorizontalPaddingRight");
        }
        get separatorHorizontalPaddingTop() {
            return this.af;
        }
        set separatorHorizontalPaddingTop(a) {
            this.af = a;
            this.g("SeparatorHorizontalPaddingTop");
        }
        get actualSeparatorHorizontalPaddingTop() {
            return this.s;
        }
        set actualSeparatorHorizontalPaddingTop(a) {
            this.s = a;
            this.g("ActualSeparatorHorizontalPaddingTop");
        }
        get separatorHorizontalPaddingBottom() {
            return this.ac;
        }
        set separatorHorizontalPaddingBottom(a) {
            this.ac = a;
            this.g("SeparatorHorizontalPaddingBottom");
        }
        get actualSeparatorHorizontalPaddingBottom() {
            return this.p;
        }
        set actualSeparatorHorizontalPaddingBottom(a) {
            this.p = a;
            this.g("ActualSeparatorHorizontalPaddingBottom");
        }
        get orientation() {
            return this.cs;
        }
        set orientation(a) {
            this.cs = a;
            this.g("Orientation");
        }
        get rowHeight() {
            return this.bg;
        }
        set rowHeight(a) {
            this.bg = a;
            this.g("RowHeight");
        }
        get highlightWidth() {
            return this.ab;
        }
        set highlightWidth(a) {
            this.ab = a;
            this.g("HighlightWidth");
        }
        get actualHighlightWidth() {
            return this.o;
        }
        set actualHighlightWidth(a) {
            this.o = a;
            this.g("ActualHighlightWidth");
        }
        get highlightRadius() {
            return this.aa;
        }
        set highlightRadius(a) {
            this.aa = a;
            this.g("HighlightRadius");
        }
        get actualHighlightRadius() {
            return this.n;
        }
        set actualHighlightRadius(a) {
            this.n = a;
            this.g("ActualHighlightRadius");
        }
        get highlightColor() {
            return this.cl;
        }
        set highlightColor(a) {
            this.cl = a;
            this.g("HighlightColor");
        }
        get actualHighlightColor() {
            return this.bu;
        }
        set actualHighlightColor(a) {
            this.bu = a;
            this.g("ActualHighlightColor");
        }
        get menuArrowStroke() {
            return this.cp;
        }
        set menuArrowStroke(a) {
            this.cp = a;
            this.g("MenuArrowStroke");
        }
        get actualMenuArrowStroke() {
            return this.by;
        }
        set actualMenuArrowStroke(a) {
            this.by = a;
            this.g("ActualMenuArrowStroke");
        }
        get uncheckedBackgroundColor() {
            return this.c2;
        }
        set uncheckedBackgroundColor(a) {
            this.c2 = a;
            this.g("UncheckedBackgroundColor");
        }
        get actualUncheckedBackgroundColor() {
            return this.b5;
        }
        set actualUncheckedBackgroundColor(a) {
            this.b5 = a;
            this.g("ActualUncheckedBackgroundColor");
        }
        get checkedBackgroundColor() {
            return this.cc;
        }
        set checkedBackgroundColor(a) {
            this.cc = a;
            this.g("CheckedBackgroundColor");
        }
        get actualCheckedBackgroundColor() {
            return this.bm;
        }
        set actualCheckedBackgroundColor(a) {
            this.bm = a;
            this.g("ActualCheckedBackgroundColor");
        }
        get uncheckedBorderColor() {
            return this.c3;
        }
        set uncheckedBorderColor(a) {
            this.c3 = a;
            this.g("UncheckedBorderColor");
        }
        get actualUncheckedBorderColor() {
            return this.b6;
        }
        set actualUncheckedBorderColor(a) {
            this.b6 = a;
            this.g("ActualUncheckedBorderColor");
        }
        get checkedBorderColor() {
            return this.cd;
        }
        set checkedBorderColor(a) {
            this.cd = a;
            this.g("CheckedBorderColor");
        }
        get actualCheckedBorderColor() {
            return this.bn;
        }
        set actualCheckedBorderColor(a) {
            this.bn = a;
            this.g("ActualCheckedBorderColor");
        }
        get dropdownDelay() {
            return this.bf;
        }
        set dropdownDelay(a) {
            this.bf = a;
            this.g("DropdownDelay");
        }
        get actualDropdownDelay() {
            return this.bd;
        }
        set actualDropdownDelay(a) {
            this.bd = a;
            this.g("ActualDropdownDelay");
        }
        get nestedActionMode() {
            return this.cq;
        }
        set nestedActionMode(a) {
            this.cq = a;
            this.g("NestedActionMode");
        }
        get contentRefreshedRef() {
            return this.ce;
        }
        set contentRefreshedRef(a) {
            this.ce = a;
            this.g("ContentRefreshedRef");
        }
        get onCommandRef() {
            return this.cr;
        }
        set onCommandRef(a) {
            this.cr = a;
            this.g("OnCommandRef");
        }
    }
    ToolPanelDescription.$t = /*@__PURE__*/ markType(ToolPanelDescription, 'ToolPanelDescription', Description.$);
    return ToolPanelDescription;
})();
