// igniteui-react-core/esm2015/lib/ToolActionRadioInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionInfoDescription } from "./ToolActionInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionRadioInfoDescription = /*@__PURE__*/ (() => {
    class ToolActionRadioInfoDescription extends ToolActionInfoDescription {
        get_type() {
            return "ToolActionRadioInfo";
        }
        constructor() {
            super();
            this.bd = false;
            this.bg = null;
        }
        get isChecked() {
            return this.bd;
        }
        set isChecked(a) {
            this.bd = a;
            this.g("IsChecked");
        }
        get channel() {
            return this.bg;
        }
        set channel(a) {
            this.bg = a;
            this.g("Channel");
        }
    }
    ToolActionRadioInfoDescription.$t = /*@__PURE__*/ markType(ToolActionRadioInfoDescription, 'ToolActionRadioInfoDescription', ToolActionInfoDescription.$);
    ToolActionRadioInfoDescription.__marshalByValue1 = true;
    ToolActionRadioInfoDescription.__marshalByValueAlias1 = "ToolActionRadioInfo";
    return ToolActionRadioInfoDescription;
})();
