// igniteui-react-core/esm2015/lib/ToolActionCheckboxListInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionInfo } from "./ToolActionInfo";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionCheckboxListInfo = /*@__PURE__*/ (() => {
    class ToolActionCheckboxListInfo extends ToolActionInfo {
        constructor() {
            super(...arguments);
            this._itemsSource = null;
            this._dataMemberPath = null;
            this._selectedMemberPath = null;
            this._showSelectAll = false;
            this._primaryKey = null;
        }
        get_d() {
            return 11;
        }
        get itemsSource() {
            return this._itemsSource;
        }
        set itemsSource(a) {
            this._itemsSource = a;
        }
        get dataMemberPath() {
            return this._dataMemberPath;
        }
        set dataMemberPath(a) {
            this._dataMemberPath = a;
        }
        get selectedMemberPath() {
            return this._selectedMemberPath;
        }
        set selectedMemberPath(a) {
            this._selectedMemberPath = a;
        }
        get showSelectAll() {
            return this._showSelectAll;
        }
        set showSelectAll(a) {
            this._showSelectAll = a;
        }
        get primaryKey() {
            return this._primaryKey;
        }
        set primaryKey(a) {
            this._primaryKey = a;
        }
    }
    ToolActionCheckboxListInfo.$t = /*@__PURE__*/ markType(ToolActionCheckboxListInfo, 'ToolActionCheckboxListInfo', ToolActionInfo.$);
    return ToolActionCheckboxListInfo;
})();
