// igniteui-react-core/esm2015/lib/ToolActionButtonPairInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionInfo } from "./ToolActionInfo";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionButtonPairInfo = /*@__PURE__*/ (() => {
    class ToolActionButtonPairInfo extends ToolActionInfo {
        get_d() {
            return 12;
        }
        constructor() {
            super();
            this._leftTitle = null;
            this._leftCommandArgument = null;
            this._leftIconName = null;
            this._leftIconCollectionName = null;
            this._isToggleDisabled = false;
            this._leftIconViewBoxLeft = 0;
            this._leftIconViewBoxTop = 0;
            this._leftIconViewBoxWidth = 0;
            this._leftIconViewBoxHeight = 0;
            this._rightTitle = null;
            this._rightCommandArgument = null;
            this._rightIconName = null;
            this._rightIconCollectionName = null;
            this._rightIconViewBoxLeft = 0;
            this._rightIconViewBoxTop = 0;
            this._rightIconViewBoxWidth = 0;
            this._rightIconViewBoxHeight = 0;
            this.leftIconViewBoxLeft = NaN;
            this.leftIconViewBoxTop = NaN;
            this.leftIconViewBoxWidth = NaN;
            this.leftIconViewBoxHeight = NaN;
            this.rightIconViewBoxLeft = NaN;
            this.rightIconViewBoxTop = NaN;
            this.rightIconViewBoxWidth = NaN;
            this.rightIconViewBoxHeight = NaN;
        }
        get leftTitle() {
            return this._leftTitle;
        }
        set leftTitle(a) {
            this._leftTitle = a;
        }
        get leftCommandArgument() {
            return this._leftCommandArgument;
        }
        set leftCommandArgument(a) {
            this._leftCommandArgument = a;
        }
        get leftIconName() {
            return this._leftIconName;
        }
        set leftIconName(a) {
            this._leftIconName = a;
        }
        get leftIconCollectionName() {
            return this._leftIconCollectionName;
        }
        set leftIconCollectionName(a) {
            this._leftIconCollectionName = a;
        }
        get isToggleDisabled() {
            return this._isToggleDisabled;
        }
        set isToggleDisabled(a) {
            this._isToggleDisabled = a;
        }
        get leftIconViewBoxLeft() {
            return this._leftIconViewBoxLeft;
        }
        set leftIconViewBoxLeft(a) {
            this._leftIconViewBoxLeft = a;
        }
        get leftIconViewBoxTop() {
            return this._leftIconViewBoxTop;
        }
        set leftIconViewBoxTop(a) {
            this._leftIconViewBoxTop = a;
        }
        get leftIconViewBoxWidth() {
            return this._leftIconViewBoxWidth;
        }
        set leftIconViewBoxWidth(a) {
            this._leftIconViewBoxWidth = a;
        }
        get leftIconViewBoxHeight() {
            return this._leftIconViewBoxHeight;
        }
        set leftIconViewBoxHeight(a) {
            this._leftIconViewBoxHeight = a;
        }
        get rightTitle() {
            return this._rightTitle;
        }
        set rightTitle(a) {
            this._rightTitle = a;
        }
        get rightCommandArgument() {
            return this._rightCommandArgument;
        }
        set rightCommandArgument(a) {
            this._rightCommandArgument = a;
        }
        get rightIconName() {
            return this._rightIconName;
        }
        set rightIconName(a) {
            this._rightIconName = a;
        }
        get rightIconCollectionName() {
            return this._rightIconCollectionName;
        }
        set rightIconCollectionName(a) {
            this._rightIconCollectionName = a;
        }
        get rightIconViewBoxLeft() {
            return this._rightIconViewBoxLeft;
        }
        set rightIconViewBoxLeft(a) {
            this._rightIconViewBoxLeft = a;
        }
        get rightIconViewBoxTop() {
            return this._rightIconViewBoxTop;
        }
        set rightIconViewBoxTop(a) {
            this._rightIconViewBoxTop = a;
        }
        get rightIconViewBoxWidth() {
            return this._rightIconViewBoxWidth;
        }
        set rightIconViewBoxWidth(a) {
            this._rightIconViewBoxWidth = a;
        }
        get rightIconViewBoxHeight() {
            return this._rightIconViewBoxHeight;
        }
        set rightIconViewBoxHeight(a) {
            this._rightIconViewBoxHeight = a;
        }
    }
    ToolActionButtonPairInfo.$t = /*@__PURE__*/ markType(ToolActionButtonPairInfo, 'ToolActionButtonPairInfo', ToolActionInfo.$);
    return ToolActionButtonPairInfo;
})();
