// igniteui-react-core/esm2015/lib/ToolPanelContextChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolPanelContextChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class ToolPanelContextChangedEventArgsDescription extends Description {
        get_type() {
            return "ToolPanelContextChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = null;
            this.i = null;
        }
        get contextKey() {
            return this.h;
        }
        set contextKey(a) {
            this.h = a;
            this.g("ContextKey");
        }
        get oldValueRef() {
            return this.j;
        }
        set oldValueRef(a) {
            this.j = a;
            this.g("OldValueRef");
        }
        get newValueRef() {
            return this.i;
        }
        set newValueRef(a) {
            this.i = a;
            this.g("NewValueRef");
        }
    }
    ToolPanelContextChangedEventArgsDescription.$t = /*@__PURE__*/ markType(ToolPanelContextChangedEventArgsDescription, 'ToolPanelContextChangedEventArgsDescription', Description.$);
    return ToolPanelContextChangedEventArgsDescription;
})();
