// igniteui-react-core/esm2015/lib/ToolActionCheckboxDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionCheckboxDescription = /*@__PURE__*/ (() => {
    class ToolActionCheckboxDescription extends ToolActionDescription {
        get_type() {
            return "ToolActionCheckbox";
        }
        constructor() {
            super();
            this.dp = false;
            this.dx = null;
            this.dt = null;
            this.dv = null;
            this.dr = null;
            this.dy = null;
            this.du = null;
            this.dw = null;
            this.ds = null;
        }
        get isChecked() {
            return this.dp;
        }
        set isChecked(a) {
            this.dp = a;
            this.g("IsChecked");
        }
        get uncheckedBackgroundColor() {
            return this.dx;
        }
        set uncheckedBackgroundColor(a) {
            this.dx = a;
            this.g("UncheckedBackgroundColor");
        }
        get actualUncheckedBackgroundColor() {
            return this.dt;
        }
        set actualUncheckedBackgroundColor(a) {
            this.dt = a;
            this.g("ActualUncheckedBackgroundColor");
        }
        get checkedBackgroundColor() {
            return this.dv;
        }
        set checkedBackgroundColor(a) {
            this.dv = a;
            this.g("CheckedBackgroundColor");
        }
        get actualCheckedBackgroundColor() {
            return this.dr;
        }
        set actualCheckedBackgroundColor(a) {
            this.dr = a;
            this.g("ActualCheckedBackgroundColor");
        }
        get uncheckedBorderColor() {
            return this.dy;
        }
        set uncheckedBorderColor(a) {
            this.dy = a;
            this.g("UncheckedBorderColor");
        }
        get actualUncheckedBorderColor() {
            return this.du;
        }
        set actualUncheckedBorderColor(a) {
            this.du = a;
            this.g("ActualUncheckedBorderColor");
        }
        get checkedBorderColor() {
            return this.dw;
        }
        set checkedBorderColor(a) {
            this.dw = a;
            this.g("CheckedBorderColor");
        }
        get actualCheckedBorderColor() {
            return this.ds;
        }
        set actualCheckedBorderColor(a) {
            this.ds = a;
            this.g("ActualCheckedBorderColor");
        }
    }
    ToolActionCheckboxDescription.$t = /*@__PURE__*/ markType(ToolActionCheckboxDescription, 'ToolActionCheckboxDescription', ToolActionDescription.$);
    return ToolActionCheckboxDescription;
})();
