// igniteui-react-core/esm2015/lib/ToolActionCheckboxListDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionCheckboxListDescription = /*@__PURE__*/ (() => {
    class ToolActionCheckboxListDescription extends ToolActionDescription {
        get_type() {
            return "ToolActionCheckboxList";
        }
        constructor() {
            super();
            this.dw = null;
            this.dy = null;
            this.dt = false;
            this.dv = null;
            this.dp = null;
            this.dq = null;
            this.dx = null;
        }
        get dataSourceRef() {
            return this.dw;
        }
        set dataSourceRef(a) {
            this.dw = a;
            this.g("DataSourceRef");
        }
        get selectedMemberPath() {
            return this.dy;
        }
        set selectedMemberPath(a) {
            this.dy = a;
            this.g("SelectedMemberPath");
        }
        get showSelectAll() {
            return this.dt;
        }
        set showSelectAll(a) {
            this.dt = a;
            this.g("ShowSelectAll");
        }
        get dataMemberPath() {
            return this.dv;
        }
        set dataMemberPath(a) {
            this.dv = a;
            this.g("DataMemberPath");
        }
        get primaryKey() {
            return this.dp;
        }
        set primaryKey(a) {
            this.dp = a;
            this.g("PrimaryKey");
        }
        get selectedKeys() {
            return this.dq;
        }
        set selectedKeys(a) {
            this.dq = a;
            this.g("SelectedKeys");
        }
        get indexType() {
            return this.dx;
        }
        set indexType(a) {
            this.dx = a;
            this.g("IndexType");
        }
    }
    ToolActionCheckboxListDescription.$t = /*@__PURE__*/ markType(ToolActionCheckboxListDescription, 'ToolActionCheckboxListDescription', ToolActionDescription.$);
    return ToolActionCheckboxListDescription;
})();
