// igniteui-react-core/esm2015/lib/ToolActionComboInfo.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionInfo } from "./ToolActionInfo";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionComboInfo = /*@__PURE__*/ (() => {
    class ToolActionComboInfo extends ToolActionInfo {
        constructor() {
            super(...arguments);
            this._itemsSource = null;
            this._displayMemberPath = null;
            this._valueMemberPath = null;
            this._selectedValues = null;
        }
        get_d() {
            return 13;
        }
        get itemsSource() {
            return this._itemsSource;
        }
        set itemsSource(a) {
            this._itemsSource = a;
        }
        get displayMemberPath() {
            return this._displayMemberPath;
        }
        set displayMemberPath(a) {
            this._displayMemberPath = a;
        }
        get valueMemberPath() {
            return this._valueMemberPath;
        }
        set valueMemberPath(a) {
            this._valueMemberPath = a;
        }
        get selectedValues() {
            return this._selectedValues;
        }
        set selectedValues(a) {
            this._selectedValues = a;
        }
    }
    ToolActionComboInfo.$t = /*@__PURE__*/ markType(ToolActionComboInfo, 'ToolActionComboInfo', ToolActionInfo.$);
    return ToolActionComboInfo;
})();
