// igniteui-react-core/esm2015/lib/ToolContextBindingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolContextBindingDescription = /*@__PURE__*/ (() => {
    class ToolContextBindingDescription extends Description {
        get_type() {
            return "ToolContextBinding";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.i = null;
            this.j = null;
            this.k = null;
        }
        get bindingMode() {
            return this.h;
        }
        set bindingMode(a) {
            this.h = a;
            this.g("BindingMode");
        }
        get contextKey() {
            return this.i;
        }
        set contextKey(a) {
            this.i = a;
            this.g("ContextKey");
        }
        get propertyName() {
            return this.j;
        }
        set propertyName(a) {
            this.j = a;
            this.g("PropertyName");
        }
        get propertyUpdatedRef() {
            return this.k;
        }
        set propertyUpdatedRef(a) {
            this.k = a;
            this.g("PropertyUpdatedRef");
        }
    }
    ToolContextBindingDescription.$t = /*@__PURE__*/ markType(ToolContextBindingDescription, 'ToolContextBindingDescription', Description.$);
    return ToolContextBindingDescription;
})();
