// igniteui-react-core/esm2015/lib/ToolActionButtonPairDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionButtonPairDescription = /*@__PURE__*/ (() => {
    class ToolActionButtonPairDescription extends ToolActionDescription {
        get_type() {
            return "ToolActionButtonPair";
        }
        constructor() {
            super();
            this.dz = false;
            this.d1 = false;
            this.dy = false;
            this.d0 = false;
            this.e1 = null;
            this.ey = null;
            this.dx = false;
            this.ez = null;
            this.ea = 0;
            this.eb = 0;
            this.ec = 0;
            this.d9 = 0;
            this.e0 = null;
            this.et = null;
            this.dp = null;
            this.e2 = null;
            this.eu = null;
            this.dq = null;
            this.d8 = 0;
            this.e3 = null;
            this.e7 = null;
            this.e4 = null;
            this.e5 = null;
            this.ef = 0;
            this.eg = 0;
            this.eh = 0;
            this.ee = 0;
            this.e6 = null;
            this.ev = null;
            this.dr = null;
            this.e8 = null;
            this.ew = null;
            this.ds = null;
            this.ed = 0;
            this.e9 = null;
            this.ex = null;
            this.d7 = 0;
        }
        get leftIsSelected() {
            return this.dz;
        }
        set leftIsSelected(a) {
            this.dz = a;
            this.g("LeftIsSelected");
        }
        get rightIsSelected() {
            return this.d1;
        }
        set rightIsSelected(a) {
            this.d1 = a;
            this.g("RightIsSelected");
        }
        get leftIsDisabled() {
            return this.dy;
        }
        set leftIsDisabled(a) {
            this.dy = a;
            this.g("LeftIsDisabled");
        }
        get rightIsDisabled() {
            return this.d0;
        }
        set rightIsDisabled(a) {
            this.d0 = a;
            this.g("RightIsDisabled");
        }
        get leftIconName() {
            return this.e1;
        }
        set leftIconName(a) {
            this.e1 = a;
            this.g("LeftIconName");
        }
        get leftCommandArgument() {
            return this.ey;
        }
        set leftCommandArgument(a) {
            this.ey = a;
            this.g("LeftCommandArgument");
        }
        get isToggleDisabled() {
            return this.dx;
        }
        set isToggleDisabled(a) {
            this.dx = a;
            this.g("IsToggleDisabled");
        }
        get leftIconCollectionName() {
            return this.ez;
        }
        set leftIconCollectionName(a) {
            this.ez = a;
            this.g("LeftIconCollectionName");
        }
        get leftIconViewBoxLeft() {
            return this.ea;
        }
        set leftIconViewBoxLeft(a) {
            this.ea = a;
            this.g("LeftIconViewBoxLeft");
        }
        get leftIconViewBoxTop() {
            return this.eb;
        }
        set leftIconViewBoxTop(a) {
            this.eb = a;
            this.g("LeftIconViewBoxTop");
        }
        get leftIconViewBoxWidth() {
            return this.ec;
        }
        set leftIconViewBoxWidth(a) {
            this.ec = a;
            this.g("LeftIconViewBoxWidth");
        }
        get leftIconViewBoxHeight() {
            return this.d9;
        }
        set leftIconViewBoxHeight(a) {
            this.d9 = a;
            this.g("LeftIconViewBoxHeight");
        }
        get leftIconFill() {
            return this.e0;
        }
        set leftIconFill(a) {
            this.e0 = a;
            this.g("LeftIconFill");
        }
        get actualLeftIconFill() {
            return this.et;
        }
        set actualLeftIconFill(a) {
            this.et = a;
            this.g("ActualLeftIconFill");
        }
        get leftIconFillColors() {
            return this.dp;
        }
        set leftIconFillColors(a) {
            this.dp = a;
            this.g("LeftIconFillColors");
        }
        get leftIconStroke() {
            return this.e2;
        }
        set leftIconStroke(a) {
            this.e2 = a;
            this.g("LeftIconStroke");
        }
        get actualLeftIconStroke() {
            return this.eu;
        }
        set actualLeftIconStroke(a) {
            this.eu = a;
            this.g("ActualLeftIconStroke");
        }
        get leftIconStrokeColors() {
            return this.dq;
        }
        set leftIconStrokeColors(a) {
            this.dq = a;
            this.g("LeftIconStrokeColors");
        }
        get leftIconStrokeWidth() {
            return this.d8;
        }
        set leftIconStrokeWidth(a) {
            this.d8 = a;
            this.g("LeftIconStrokeWidth");
        }
        get leftTitle() {
            return this.e3;
        }
        set leftTitle(a) {
            this.e3 = a;
            this.g("LeftTitle");
        }
        get rightIconName() {
            return this.e7;
        }
        set rightIconName(a) {
            this.e7 = a;
            this.g("RightIconName");
        }
        get rightCommandArgument() {
            return this.e4;
        }
        set rightCommandArgument(a) {
            this.e4 = a;
            this.g("RightCommandArgument");
        }
        get rightIconCollectionName() {
            return this.e5;
        }
        set rightIconCollectionName(a) {
            this.e5 = a;
            this.g("RightIconCollectionName");
        }
        get rightIconViewBoxLeft() {
            return this.ef;
        }
        set rightIconViewBoxLeft(a) {
            this.ef = a;
            this.g("RightIconViewBoxLeft");
        }
        get rightIconViewBoxTop() {
            return this.eg;
        }
        set rightIconViewBoxTop(a) {
            this.eg = a;
            this.g("RightIconViewBoxTop");
        }
        get rightIconViewBoxWidth() {
            return this.eh;
        }
        set rightIconViewBoxWidth(a) {
            this.eh = a;
            this.g("RightIconViewBoxWidth");
        }
        get rightIconViewBoxHeight() {
            return this.ee;
        }
        set rightIconViewBoxHeight(a) {
            this.ee = a;
            this.g("RightIconViewBoxHeight");
        }
        get rightIconFill() {
            return this.e6;
        }
        set rightIconFill(a) {
            this.e6 = a;
            this.g("RightIconFill");
        }
        get actualRightIconFill() {
            return this.ev;
        }
        set actualRightIconFill(a) {
            this.ev = a;
            this.g("ActualRightIconFill");
        }
        get rightIconFillColors() {
            return this.dr;
        }
        set rightIconFillColors(a) {
            this.dr = a;
            this.g("RightIconFillColors");
        }
        get rightIconStroke() {
            return this.e8;
        }
        set rightIconStroke(a) {
            this.e8 = a;
            this.g("RightIconStroke");
        }
        get actualRightIconStroke() {
            return this.ew;
        }
        set actualRightIconStroke(a) {
            this.ew = a;
            this.g("ActualRightIconStroke");
        }
        get rightIconStrokeColors() {
            return this.ds;
        }
        set rightIconStrokeColors(a) {
            this.ds = a;
            this.g("RightIconStrokeColors");
        }
        get rightIconStrokeWidth() {
            return this.ed;
        }
        set rightIconStrokeWidth(a) {
            this.ed = a;
            this.g("RightIconStrokeWidth");
        }
        get rightTitle() {
            return this.e9;
        }
        set rightTitle(a) {
            this.e9 = a;
            this.g("RightTitle");
        }
        get displayType() {
            return this.ex;
        }
        set displayType(a) {
            this.ex = a;
            this.g("DisplayType");
        }
        get cornerRadius() {
            return this.d7;
        }
        set cornerRadius(a) {
            this.d7 = a;
            this.g("CornerRadius");
        }
    }
    ToolActionButtonPairDescription.$t = /*@__PURE__*/ markType(ToolActionButtonPairDescription, 'ToolActionButtonPairDescription', ToolActionDescription.$);
    return ToolActionButtonPairDescription;
})();
