// igniteui-react-core/esm2015/lib/ToolActionEventDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionEventDetailDescription = /*@__PURE__*/ (() => {
    class ToolActionEventDetailDescription extends Description {
        get_type() {
            return "ToolActionEventDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.p = null;
            this.q = null;
            this.i = false;
            this.h = false;
            this.n = 0;
            this.l = new Date();
            this.s = null;
            this.r = null;
        }
        get actionId() {
            return this.p;
        }
        set actionId(a) {
            this.p = a;
            this.g("ActionId");
        }
        get actionType() {
            return this.q;
        }
        set actionType(a) {
            this.q = a;
            this.g("ActionType");
        }
        get isModified() {
            return this.i;
        }
        set isModified(a) {
            this.i = a;
            this.g("IsModified");
        }
        get boolValue() {
            return this.h;
        }
        set boolValue(a) {
            this.h = a;
            this.g("BoolValue");
        }
        get numberValue() {
            return this.n;
        }
        set numberValue(a) {
            this.n = a;
            this.g("NumberValue");
        }
        get dateTimeValue() {
            return this.l;
        }
        set dateTimeValue(a) {
            this.l = a;
            this.g("DateTimeValue");
        }
        get untypedValueRef() {
            return this.s;
        }
        set untypedValueRef(a) {
            this.s = a;
            this.g("UntypedValueRef");
        }
        get stringValue() {
            return this.r;
        }
        set stringValue(a) {
            this.r = a;
            this.g("StringValue");
        }
    }
    ToolActionEventDetailDescription.$t = /*@__PURE__*/ markType(ToolActionEventDetailDescription, 'ToolActionEventDetailDescription', Description.$);
    return ToolActionEventDetailDescription;
})();
