// igniteui-react-core/esm2015/lib/ToolActionRadioDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionRadioDescription = /*@__PURE__*/ (() => {
    class ToolActionRadioDescription extends ToolActionDescription {
        get_type() {
            return "ToolActionRadio";
        }
        constructor() {
            super();
            this.dx = null;
            this.dp = false;
            this.dq = false;
            this.d0 = null;
            this.dv = null;
            this.dy = null;
            this.dt = null;
            this.d1 = null;
            this.dw = null;
            this.dz = null;
            this.du = null;
        }
        get channel() {
            return this.dx;
        }
        set channel(a) {
            this.dx = a;
            this.g("Channel");
        }
        get isChecked() {
            return this.dp;
        }
        set isChecked(a) {
            this.dp = a;
            this.g("IsChecked");
        }
        get isManual() {
            return this.dq;
        }
        set isManual(a) {
            this.dq = a;
            this.g("IsManual");
        }
        get uncheckedBackgroundColor() {
            return this.d0;
        }
        set uncheckedBackgroundColor(a) {
            this.d0 = a;
            this.g("UncheckedBackgroundColor");
        }
        get actualUncheckedBackgroundColor() {
            return this.dv;
        }
        set actualUncheckedBackgroundColor(a) {
            this.dv = a;
            this.g("ActualUncheckedBackgroundColor");
        }
        get checkedBackgroundColor() {
            return this.dy;
        }
        set checkedBackgroundColor(a) {
            this.dy = a;
            this.g("CheckedBackgroundColor");
        }
        get actualCheckedBackgroundColor() {
            return this.dt;
        }
        set actualCheckedBackgroundColor(a) {
            this.dt = a;
            this.g("ActualCheckedBackgroundColor");
        }
        get uncheckedBorderColor() {
            return this.d1;
        }
        set uncheckedBorderColor(a) {
            this.d1 = a;
            this.g("UncheckedBorderColor");
        }
        get actualUncheckedBorderColor() {
            return this.dw;
        }
        set actualUncheckedBorderColor(a) {
            this.dw = a;
            this.g("ActualUncheckedBorderColor");
        }
        get checkedBorderColor() {
            return this.dz;
        }
        set checkedBorderColor(a) {
            this.dz = a;
            this.g("CheckedBorderColor");
        }
        get actualCheckedBorderColor() {
            return this.du;
        }
        set actualCheckedBorderColor(a) {
            this.du = a;
            this.g("ActualCheckedBorderColor");
        }
    }
    ToolActionRadioDescription.$t = /*@__PURE__*/ markType(ToolActionRadioDescription, 'ToolActionRadioDescription', ToolActionDescription.$);
    return ToolActionRadioDescription;
})();
