// igniteui-react-core/esm2015/lib/ToolActionInfoDensity.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
export var ToolActionInfoDensity = /*@__PURE__*/ (function (ToolActionInfoDensity) {
    ToolActionInfoDensity[ToolActionInfoDensity["Auto"] = 0] = "Auto";
    ToolActionInfoDensity[ToolActionInfoDensity["Comfortable"] = 1] = "Comfortable";
    ToolActionInfoDensity[ToolActionInfoDensity["Cosy"] = 2] = "Cosy";
    ToolActionInfoDensity[ToolActionInfoDensity["Compact"] = 3] = "Compact";
    ToolActionInfoDensity[ToolActionInfoDensity["Minimal"] = 4] = "Minimal";
    return ToolActionInfoDensity;
})({});
/**
 * @hidden
 */
export let ToolActionInfoDensity_$type = /*@__PURE__*/ markEnum('ToolActionInfoDensity', 'Auto,0|Comfortable,1|Cosy,2|Compact,3|Minimal,4');
