// igniteui-react-core/esm2015/lib/ToolContextBindingInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolContextBindingInfoDescription = /*@__PURE__*/ (() => {
    class ToolContextBindingInfoDescription extends Description {
        get_type() {
            return "ToolContextBindingInfo";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = null;
            this.l = null;
            this.j = null;
        }
        get contextKey() {
            return this.k;
        }
        set contextKey(a) {
            this.k = a;
            this.g("ContextKey");
        }
        get propertyName() {
            return this.l;
        }
        set propertyName(a) {
            this.l = a;
            this.g("PropertyName");
        }
        get bindingMode() {
            return this.j;
        }
        set bindingMode(a) {
            this.j = a;
            this.g("BindingMode");
        }
    }
    ToolContextBindingInfoDescription.$t = /*@__PURE__*/ markType(ToolContextBindingInfoDescription, 'ToolContextBindingInfoDescription', Description.$);
    ToolContextBindingInfoDescription.__marshalByValue = true;
    ToolContextBindingInfoDescription.__marshalByValueAlias = "ToolContextBindingInfo";
    return ToolContextBindingInfoDescription;
})();
