// igniteui-react-core/esm2015/lib/ToolActionComboInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionInfoDescription } from "./ToolActionInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionComboInfoDescription = /*@__PURE__*/ (() => {
    class ToolActionComboInfoDescription extends ToolActionInfoDescription {
        get_type() {
            return "ToolActionComboInfo";
        }
        constructor() {
            super();
            this.be = null;
            this.bf = null;
            this.bh = null;
            this.bg = null;
        }
        get dataSourceRef() {
            return this.be;
        }
        set dataSourceRef(a) {
            this.be = a;
            this.g("DataSourceRef");
        }
        get displayMemberPath() {
            return this.bf;
        }
        set displayMemberPath(a) {
            this.bf = a;
            this.g("DisplayMemberPath");
        }
        get valueMemberPath() {
            return this.bh;
        }
        set valueMemberPath(a) {
            this.bh = a;
            this.g("ValueMemberPath");
        }
        get selectedValuesRef() {
            return this.bg;
        }
        set selectedValuesRef(a) {
            this.bg = a;
            this.g("SelectedValuesRef");
        }
    }
    ToolActionComboInfoDescription.$t = /*@__PURE__*/ markType(ToolActionComboInfoDescription, 'ToolActionComboInfoDescription', ToolActionInfoDescription.$);
    ToolActionComboInfoDescription.__marshalByValue1 = true;
    ToolActionComboInfoDescription.__marshalByValueAlias1 = "ToolActionComboInfo";
    return ToolActionComboInfoDescription;
})();
