// igniteui-react-core/esm2015/lib/TreemapNodeStyleDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let TreemapNodeStyleDescription = /*@__PURE__*/ (() => {
    class TreemapNodeStyleDescription extends Description {
        get_type() {
            return "TreemapNodeStyle";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.aj = null;
            this.ai = null;
            this.al = null;
            this.am = null;
            this.aq = null;
            this.ak = null;
            this.m = 0;
            this.o = 0;
            this.n = 0;
            this.l = 0;
            this.q = 0;
            this.s = 0;
            this.r = 0;
            this.p = 0;
            this.k = 0;
            this.an = null;
            this.ao = null;
            this.ah = null;
            this.ap = null;
            this.u = 0;
            this.t = 0;
            this.j = 0;
            this.h = false;
        }
        get headerHoverBackground() {
            return this.aj;
        }
        set headerHoverBackground(a) {
            this.aj = a;
            this.g("HeaderHoverBackground");
        }
        get headerBackground() {
            return this.ai;
        }
        set headerBackground(a) {
            this.ai = a;
            this.g("HeaderBackground");
        }
        get headerTextColor() {
            return this.al;
        }
        set headerTextColor(a) {
            this.al = a;
            this.g("HeaderTextColor");
        }
        get label() {
            return this.am;
        }
        set label(a) {
            this.am = a;
            this.g("Label");
        }
        get textColor() {
            return this.aq;
        }
        set textColor(a) {
            this.aq = a;
            this.g("TextColor");
        }
        get headerHoverTextColor() {
            return this.ak;
        }
        set headerHoverTextColor(a) {
            this.ak = a;
            this.g("HeaderHoverTextColor");
        }
        get headerLabelLeftMargin() {
            return this.m;
        }
        set headerLabelLeftMargin(a) {
            this.m = a;
            this.g("HeaderLabelLeftMargin");
        }
        get headerLabelTopMargin() {
            return this.o;
        }
        set headerLabelTopMargin(a) {
            this.o = a;
            this.g("HeaderLabelTopMargin");
        }
        get headerLabelRightMargin() {
            return this.n;
        }
        set headerLabelRightMargin(a) {
            this.n = a;
            this.g("HeaderLabelRightMargin");
        }
        get headerLabelBottomMargin() {
            return this.l;
        }
        set headerLabelBottomMargin(a) {
            this.l = a;
            this.g("HeaderLabelBottomMargin");
        }
        get labelLeftMargin() {
            return this.q;
        }
        set labelLeftMargin(a) {
            this.q = a;
            this.g("LabelLeftMargin");
        }
        get labelTopMargin() {
            return this.s;
        }
        set labelTopMargin(a) {
            this.s = a;
            this.g("LabelTopMargin");
        }
        get labelRightMargin() {
            return this.r;
        }
        set labelRightMargin(a) {
            this.r = a;
            this.g("LabelRightMargin");
        }
        get labelBottomMargin() {
            return this.p;
        }
        set labelBottomMargin(a) {
            this.p = a;
            this.g("LabelBottomMargin");
        }
        get headerHeight() {
            return this.k;
        }
        set headerHeight(a) {
            this.k = a;
            this.g("HeaderHeight");
        }
        get labelHorizontalAlignment() {
            return this.an;
        }
        set labelHorizontalAlignment(a) {
            this.an = a;
            this.g("LabelHorizontalAlignment");
        }
        get labelVerticalAlignment() {
            return this.ao;
        }
        set labelVerticalAlignment(a) {
            this.ao = a;
            this.g("LabelVerticalAlignment");
        }
        get fill() {
            return this.ah;
        }
        set fill(a) {
            this.ah = a;
            this.g("Fill");
        }
        get outline() {
            return this.ap;
        }
        set outline(a) {
            this.ap = a;
            this.g("Outline");
        }
        get strokeThickness() {
            return this.u;
        }
        set strokeThickness(a) {
            this.u = a;
            this.g("StrokeThickness");
        }
        get opacity() {
            return this.t;
        }
        set opacity(a) {
            this.t = a;
            this.g("Opacity");
        }
        get fadeOpacity() {
            return this.j;
        }
        set fadeOpacity(a) {
            this.j = a;
            this.g("FadeOpacity");
        }
        get highlightingHandled() {
            return this.h;
        }
        set highlightingHandled(a) {
            this.h = a;
            this.g("HighlightingHandled");
        }
    }
    TreemapNodeStyleDescription.$t = /*@__PURE__*/ markType(TreemapNodeStyleDescription, 'TreemapNodeStyleDescription', Description.$);
    return TreemapNodeStyleDescription;
})();
