// igniteui-react-core/esm2015/lib/TreemapNodeStylingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let TreemapNodeStylingEventArgsDescription = /*@__PURE__*/ (() => {
    class TreemapNodeStylingEventArgsDescription extends Description {
        get_type() {
            return "TreemapNodeStylingEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.v = 0;
            this.t = 0;
            this.ab = null;
            this.ac = null;
            this.ae = null;
            this.ad = null;
            this.s = 0;
            this.af = null;
            this.r = 0;
            this.n = false;
            this.j = null;
            this.h = null;
            this.l = false;
            this.m = false;
            this.u = 0;
        }
        get value() {
            return this.v;
        }
        set value(a) {
            this.v = a;
            this.g("Value");
        }
        get sum() {
            return this.t;
        }
        set sum(a) {
            this.t = a;
            this.g("Sum");
        }
        get customValueRef() {
            return this.ab;
        }
        set customValueRef(a) {
            this.ab = a;
            this.g("CustomValueRef");
        }
        get itemRef() {
            return this.ac;
        }
        set itemRef(a) {
            this.ac = a;
            this.g("ItemRef");
        }
        get parentItemRef() {
            return this.ae;
        }
        set parentItemRef(a) {
            this.ae = a;
            this.g("ParentItemRef");
        }
        get label() {
            return this.ad;
        }
        set label(a) {
            this.ad = a;
            this.g("Label");
        }
        get parentValue() {
            return this.s;
        }
        set parentValue(a) {
            this.s = a;
            this.g("ParentValue");
        }
        get parentLabel() {
            return this.af;
        }
        set parentLabel(a) {
            this.af = a;
            this.g("ParentLabel");
        }
        get parentSum() {
            return this.r;
        }
        set parentSum(a) {
            this.r = a;
            this.g("ParentSum");
        }
        get isParent() {
            return this.n;
        }
        set isParent(a) {
            this.n = a;
            this.g("IsParent");
        }
        get style() {
            return this.j;
        }
        set style(a) {
            this.j = a;
            this.g("Style");
        }
        get highlightingInfo() {
            return this.h;
        }
        set highlightingInfo(a) {
            this.h = a;
            this.g("HighlightingInfo");
        }
        get highlightingHandled() {
            return this.l;
        }
        set highlightingHandled(a) {
            this.l = a;
            this.g("HighlightingHandled");
        }
        get isHighlightInProgress() {
            return this.m;
        }
        set isHighlightInProgress(a) {
            this.m = a;
            this.g("IsHighlightInProgress");
        }
        get totalHighlightProgress() {
            return this.u;
        }
        set totalHighlightProgress(a) {
            this.u = a;
            this.g("TotalHighlightProgress");
        }
    }
    TreemapNodeStylingEventArgsDescription.$t = /*@__PURE__*/ markType(TreemapNodeStylingEventArgsDescription, 'TreemapNodeStylingEventArgsDescription', Description.$);
    return TreemapNodeStylingEventArgsDescription;
})();
