// igniteui-react-core/esm2015/lib/TreemapDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let TreemapDescription = /*@__PURE__*/ (() => {
    class TreemapDescription extends Description {
        get_type() {
            return "Treemap";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.da = null;
            this.c8 = null;
            this.b1 = null;
            this.c7 = null;
            this.cs = null;
            this.b2 = null;
            this.cp = null;
            this.b4 = null;
            this.cl = null;
            this.cm = null;
            this.y = 0;
            this.cn = null;
            this.h = null;
            this.i = null;
            this.b5 = null;
            this.l = false;
            this.bt = 0;
            this.q = 0;
            this.p = 0;
            this.cu = null;
            this.cr = null;
            this.c6 = null;
            this.cg = null;
            this.ca = null;
            this.c5 = null;
            this.cj = null;
            this.cb = null;
            this.c9 = null;
            this.b3 = null;
            this.ci = null;
            this.ch = null;
            this.b7 = null;
            this.r = 0;
            this.b8 = null;
            this.b9 = null;
            this.cd = null;
            this.s = 0;
            this.ce = null;
            this.cf = null;
            this.cc = null;
            this.v = 0;
            this.x = 0;
            this.w = 0;
            this.u = 0;
            this.ai = 0;
            this.ak = 0;
            this.aj = 0;
            this.ah = 0;
            this.ac = 0;
            this.ae = 0;
            this.ad = 0;
            this.ab = 0;
            this.an = 0;
            this.ar = 0;
            this.ap = 0;
            this.al = 0;
            this.ao = 0;
            this.as = 0;
            this.aq = 0;
            this.am = 0;
            this.cq = null;
            this.ct = null;
            this.t = 0;
            this.c4 = null;
            this.au = 0;
            this.ag = 0;
            this.cv = null;
            this.cw = null;
            this.af = 0;
            this.bv = 0;
            this.at = 0;
            this.o = 0;
            this.aa = 0;
            this.n = 0;
            this.bu = 0;
            this.co = null;
            this.bz = null;
            this.z = 0;
            this.b6 = null;
            this.c3 = null;
            this.c2 = null;
            this.cz = null;
            this.cx = null;
            this.cy = null;
            this.c0 = null;
            this.c1 = null;
            this.db = null;
            this.ck = null;
            this.b0 = null;
        }
        get valueMemberPath() {
            return this.da;
        }
        set valueMemberPath(a) {
            this.da = a;
            this.g("ValueMemberPath");
        }
        get rootTitle() {
            return this.c8;
        }
        set rootTitle(a) {
            this.c8 = a;
            this.g("RootTitle");
        }
        get breadcrumbSequence() {
            return this.b1;
        }
        set breadcrumbSequence(a) {
            this.b1 = a;
            this.g("BreadcrumbSequence");
        }
        get parentIdMemberPath() {
            return this.c7;
        }
        set parentIdMemberPath(a) {
            this.c7 = a;
            this.g("ParentIdMemberPath");
        }
        get labelMemberPath() {
            return this.cs;
        }
        set labelMemberPath(a) {
            this.cs = a;
            this.g("LabelMemberPath");
        }
        get customValueMemberPath() {
            return this.b2;
        }
        set customValueMemberPath(a) {
            this.b2 = a;
            this.g("CustomValueMemberPath");
        }
        get idMemberPath() {
            return this.cp;
        }
        set idMemberPath(a) {
            this.cp = a;
            this.g("IdMemberPath");
        }
        get dataSourceRef() {
            return this.b4;
        }
        set dataSourceRef(a) {
            this.b4 = a;
            this.g("DataSourceRef");
        }
        get highlightedDataSourceRef() {
            return this.cl;
        }
        set highlightedDataSourceRef(a) {
            this.cl = a;
            this.g("HighlightedDataSourceRef");
        }
        get highlightedValueMemberPath() {
            return this.cm;
        }
        set highlightedValueMemberPath(a) {
            this.cm = a;
            this.g("HighlightedValueMemberPath");
        }
        get highlightedValueOpacity() {
            return this.y;
        }
        set highlightedValueOpacity(a) {
            this.y = a;
            this.g("HighlightedValueOpacity");
        }
        get highlightedValuesDisplayMode() {
            return this.cn;
        }
        set highlightedValuesDisplayMode(a) {
            this.cn = a;
            this.g("HighlightedValuesDisplayMode");
        }
        get fillBrushes() {
            return this.h;
        }
        set fillBrushes(a) {
            this.h = a;
            this.g("FillBrushes");
        }
        get styleMappings() {
            return this.i;
        }
        set styleMappings(a) {
            this.i = a;
            this.g("StyleMappings");
        }
        get fillScaleMode() {
            return this.b5;
        }
        set fillScaleMode(a) {
            this.b5 = a;
            this.g("FillScaleMode");
        }
        get isFillScaleLogarithmic() {
            return this.l;
        }
        set isFillScaleLogarithmic(a) {
            this.l = a;
            this.g("IsFillScaleLogarithmic");
        }
        get fillScaleLogarithmBase() {
            return this.bt;
        }
        set fillScaleLogarithmBase(a) {
            this.bt = a;
            this.g("FillScaleLogarithmBase");
        }
        get fillScaleMinimumValue() {
            return this.q;
        }
        set fillScaleMinimumValue(a) {
            this.q = a;
            this.g("FillScaleMinimumValue");
        }
        get fillScaleMaximumValue() {
            return this.p;
        }
        set fillScaleMaximumValue(a) {
            this.p = a;
            this.g("FillScaleMaximumValue");
        }
        get labelVerticalFitMode() {
            return this.cu;
        }
        set labelVerticalFitMode(a) {
            this.cu = a;
            this.g("LabelVerticalFitMode");
        }
        get labelHorizontalFitMode() {
            return this.cr;
        }
        set labelHorizontalFitMode(a) {
            this.cr = a;
            this.g("LabelHorizontalFitMode");
        }
        get overlayHeaderHoverBackground() {
            return this.c6;
        }
        set overlayHeaderHoverBackground(a) {
            this.c6 = a;
            this.g("OverlayHeaderHoverBackground");
        }
        get headerHoverBackground() {
            return this.cg;
        }
        set headerHoverBackground(a) {
            this.cg = a;
            this.g("HeaderHoverBackground");
        }
        get headerBackground() {
            return this.ca;
        }
        set headerBackground(a) {
            this.ca = a;
            this.g("HeaderBackground");
        }
        get overlayHeaderBackground() {
            return this.c5;
        }
        set overlayHeaderBackground(a) {
            this.c5 = a;
            this.g("OverlayHeaderBackground");
        }
        get headerTextColor() {
            return this.cj;
        }
        set headerTextColor(a) {
            this.cj = a;
            this.g("HeaderTextColor");
        }
        get headerDarkTextColor() {
            return this.cb;
        }
        set headerDarkTextColor(a) {
            this.cb = a;
            this.g("HeaderDarkTextColor");
        }
        get textColor() {
            return this.c9;
        }
        set textColor(a) {
            this.c9 = a;
            this.g("TextColor");
        }
        get darkTextColor() {
            return this.b3;
        }
        set darkTextColor(a) {
            this.b3 = a;
            this.g("DarkTextColor");
        }
        get headerHoverTextColor() {
            return this.ci;
        }
        set headerHoverTextColor(a) {
            this.ci = a;
            this.g("HeaderHoverTextColor");
        }
        get headerHoverDarkTextColor() {
            return this.ch;
        }
        set headerHoverDarkTextColor(a) {
            this.ch = a;
            this.g("HeaderHoverDarkTextColor");
        }
        get fontFamily() {
            return this.b7;
        }
        set fontFamily(a) {
            this.b7 = a;
            this.g("FontFamily");
        }
        get fontSize() {
            return this.r;
        }
        set fontSize(a) {
            this.r = a;
            this.g("FontSize");
        }
        get fontStyle() {
            return this.b8;
        }
        set fontStyle(a) {
            this.b8 = a;
            this.g("FontStyle");
        }
        get fontWeight() {
            return this.b9;
        }
        set fontWeight(a) {
            this.b9 = a;
            this.g("FontWeight");
        }
        get headerFontFamily() {
            return this.cd;
        }
        set headerFontFamily(a) {
            this.cd = a;
            this.g("HeaderFontFamily");
        }
        get headerFontSize() {
            return this.s;
        }
        set headerFontSize(a) {
            this.s = a;
            this.g("HeaderFontSize");
        }
        get headerFontStyle() {
            return this.ce;
        }
        set headerFontStyle(a) {
            this.ce = a;
            this.g("HeaderFontStyle");
        }
        get headerFontWeight() {
            return this.cf;
        }
        set headerFontWeight(a) {
            this.cf = a;
            this.g("HeaderFontWeight");
        }
        get headerDisplayMode() {
            return this.cc;
        }
        set headerDisplayMode(a) {
            this.cc = a;
            this.g("HeaderDisplayMode");
        }
        get headerLabelLeftMargin() {
            return this.v;
        }
        set headerLabelLeftMargin(a) {
            this.v = a;
            this.g("HeaderLabelLeftMargin");
        }
        get headerLabelTopMargin() {
            return this.x;
        }
        set headerLabelTopMargin(a) {
            this.x = a;
            this.g("HeaderLabelTopMargin");
        }
        get headerLabelRightMargin() {
            return this.w;
        }
        set headerLabelRightMargin(a) {
            this.w = a;
            this.g("HeaderLabelRightMargin");
        }
        get headerLabelBottomMargin() {
            return this.u;
        }
        set headerLabelBottomMargin(a) {
            this.u = a;
            this.g("HeaderLabelBottomMargin");
        }
        get overlayHeaderLabelLeftMargin() {
            return this.ai;
        }
        set overlayHeaderLabelLeftMargin(a) {
            this.ai = a;
            this.g("OverlayHeaderLabelLeftMargin");
        }
        get overlayHeaderLabelTopMargin() {
            return this.ak;
        }
        set overlayHeaderLabelTopMargin(a) {
            this.ak = a;
            this.g("OverlayHeaderLabelTopMargin");
        }
        get overlayHeaderLabelRightMargin() {
            return this.aj;
        }
        set overlayHeaderLabelRightMargin(a) {
            this.aj = a;
            this.g("OverlayHeaderLabelRightMargin");
        }
        get overlayHeaderLabelBottomMargin() {
            return this.ah;
        }
        set overlayHeaderLabelBottomMargin(a) {
            this.ah = a;
            this.g("OverlayHeaderLabelBottomMargin");
        }
        get labelLeftMargin() {
            return this.ac;
        }
        set labelLeftMargin(a) {
            this.ac = a;
            this.g("LabelLeftMargin");
        }
        get labelTopMargin() {
            return this.ae;
        }
        set labelTopMargin(a) {
            this.ae = a;
            this.g("LabelTopMargin");
        }
        get labelRightMargin() {
            return this.ad;
        }
        set labelRightMargin(a) {
            this.ad = a;
            this.g("LabelRightMargin");
        }
        get labelBottomMargin() {
            return this.ab;
        }
        set labelBottomMargin(a) {
            this.ab = a;
            this.g("LabelBottomMargin");
        }
        get parentNodeLeftMargin() {
            return this.an;
        }
        set parentNodeLeftMargin(a) {
            this.an = a;
            this.g("ParentNodeLeftMargin");
        }
        get parentNodeTopMargin() {
            return this.ar;
        }
        set parentNodeTopMargin(a) {
            this.ar = a;
            this.g("ParentNodeTopMargin");
        }
        get parentNodeRightMargin() {
            return this.ap;
        }
        set parentNodeRightMargin(a) {
            this.ap = a;
            this.g("ParentNodeRightMargin");
        }
        get parentNodeBottomMargin() {
            return this.al;
        }
        set parentNodeBottomMargin(a) {
            this.al = a;
            this.g("ParentNodeBottomMargin");
        }
        get parentNodeLeftPadding() {
            return this.ao;
        }
        set parentNodeLeftPadding(a) {
            this.ao = a;
            this.g("ParentNodeLeftPadding");
        }
        get parentNodeTopPadding() {
            return this.as;
        }
        set parentNodeTopPadding(a) {
            this.as = a;
            this.g("ParentNodeTopPadding");
        }
        get parentNodeRightPadding() {
            return this.aq;
        }
        set parentNodeRightPadding(a) {
            this.aq = a;
            this.g("ParentNodeRightPadding");
        }
        get parentNodeBottomPadding() {
            return this.am;
        }
        set parentNodeBottomPadding(a) {
            this.am = a;
            this.g("ParentNodeBottomPadding");
        }
        get labelHorizontalAlignment() {
            return this.cq;
        }
        set labelHorizontalAlignment(a) {
            this.cq = a;
            this.g("LabelHorizontalAlignment");
        }
        get labelVerticalAlignment() {
            return this.ct;
        }
        set labelVerticalAlignment(a) {
            this.ct = a;
            this.g("LabelVerticalAlignment");
        }
        get headerHeight() {
            return this.t;
        }
        set headerHeight(a) {
            this.t = a;
            this.g("HeaderHeight");
        }
        get outline() {
            return this.c4;
        }
        set outline(a) {
            this.c4 = a;
            this.g("Outline");
        }
        get strokeThickness() {
            return this.au;
        }
        set strokeThickness(a) {
            this.au = a;
            this.g("StrokeThickness");
        }
        get nodeOpacity() {
            return this.ag;
        }
        set nodeOpacity(a) {
            this.ag = a;
            this.g("NodeOpacity");
        }
        get layoutOrientation() {
            return this.cv;
        }
        set layoutOrientation(a) {
            this.cv = a;
            this.g("LayoutOrientation");
        }
        get layoutType() {
            return this.cw;
        }
        set layoutType(a) {
            this.cw = a;
            this.g("LayoutType");
        }
        get minimumDisplaySize() {
            return this.af;
        }
        set minimumDisplaySize(a) {
            this.af = a;
            this.g("MinimumDisplaySize");
        }
        get transitionDuration() {
            return this.bv;
        }
        set transitionDuration(a) {
            this.bv = a;
            this.g("TransitionDuration");
        }
        get pixelScalingRatio() {
            return this.at;
        }
        set pixelScalingRatio(a) {
            this.at = a;
            this.g("PixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.o;
        }
        set actualPixelScalingRatio(a) {
            this.o = a;
            this.g("ActualPixelScalingRatio");
        }
        get interactionPixelScalingRatio() {
            return this.aa;
        }
        set interactionPixelScalingRatio(a) {
            this.aa = a;
            this.g("InteractionPixelScalingRatio");
        }
        get actualInteractionPixelScalingRatio() {
            return this.n;
        }
        set actualInteractionPixelScalingRatio(a) {
            this.n = a;
            this.g("ActualInteractionPixelScalingRatio");
        }
        get highlightingTransitionDuration() {
            return this.bu;
        }
        set highlightingTransitionDuration(a) {
            this.bu = a;
            this.g("HighlightingTransitionDuration");
        }
        get highlightingMode() {
            return this.co;
        }
        set highlightingMode(a) {
            this.co = a;
            this.g("HighlightingMode");
        }
        get actualHighlightingMode() {
            return this.bz;
        }
        set actualHighlightingMode(a) {
            this.bz = a;
            this.g("ActualHighlightingMode");
        }
        get highlightingFadeOpacity() {
            return this.z;
        }
        set highlightingFadeOpacity(a) {
            this.z = a;
            this.g("HighlightingFadeOpacity");
        }
        get focusItemRef() {
            return this.b6;
        }
        set focusItemRef(a) {
            this.b6 = a;
            this.g("FocusItemRef");
        }
        get nodeStylingRef() {
            return this.c3;
        }
        set nodeStylingRef(a) {
            this.c3 = a;
            this.g("NodeStylingRef");
        }
        get nodeRenderStylingRef() {
            return this.c2;
        }
        set nodeRenderStylingRef(a) {
            this.c2 = a;
            this.g("NodeRenderStylingRef");
        }
        get nodePointerOverRef() {
            return this.cz;
        }
        set nodePointerOverRef(a) {
            this.cz = a;
            this.g("NodePointerOverRef");
        }
        get nodePointerEnterRef() {
            return this.cx;
        }
        set nodePointerEnterRef(a) {
            this.cx = a;
            this.g("NodePointerEnterRef");
        }
        get nodePointerLeaveRef() {
            return this.cy;
        }
        set nodePointerLeaveRef(a) {
            this.cy = a;
            this.g("NodePointerLeaveRef");
        }
        get nodePointerPressedRef() {
            return this.c0;
        }
        set nodePointerPressedRef(a) {
            this.c0 = a;
            this.g("NodePointerPressedRef");
        }
        get nodePointerReleasedRef() {
            return this.c1;
        }
        set nodePointerReleasedRef(a) {
            this.c1 = a;
            this.g("NodePointerReleasedRef");
        }
        get width() {
            return this.db;
        }
        set width(a) {
            this.db = a;
            this.g("Width");
        }
        get height() {
            return this.ck;
        }
        set height(a) {
            this.ck = a;
            this.g("Height");
        }
        get background() {
            return this.b0;
        }
        set background(a) {
            this.b0 = a;
            this.g("Background");
        }
    }
    TreemapDescription.$t = /*@__PURE__*/ markType(TreemapDescription, 'TreemapDescription', Description.$);
    return TreemapDescription;
})();
