// igniteui-react-core/esm2015/lib/TriangulatorContext.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let TriangulatorContext = /*@__PURE__*/ (() => {
    class TriangulatorContext extends Base {
        constructor() {
            super(...arguments);
            this.j = 0;
            this.d = null;
            this.e = null;
            this.i = null;
            this.n = null;
            this.o = null;
            this.p = null;
            this.g = null;
            this.f = null;
            this.h = null;
            this.a = null;
            this.b = null;
            this.m = 0;
            this.l = 0;
            this.k = 0;
            this.c = false;
        }
    }
    TriangulatorContext.$t = /*@__PURE__*/ markType(TriangulatorContext, 'TriangulatorContext');
    return TriangulatorContext;
})();
