// igniteui-react-core/esm2015/lib/TypeDescriptionPlatform.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * @hidden
 */
export var TypeDescriptionPlatform = /*@__PURE__*/ (function (TypeDescriptionPlatform) {
    TypeDescriptionPlatform[TypeDescriptionPlatform["WPF"] = 0] = "WPF";
    TypeDescriptionPlatform[TypeDescriptionPlatform["JQuery"] = 1] = "JQuery";
    TypeDescriptionPlatform[TypeDescriptionPlatform["React"] = 2] = "React";
    TypeDescriptionPlatform[TypeDescriptionPlatform["Angular"] = 3] = "Angular";
    TypeDescriptionPlatform[TypeDescriptionPlatform["WebComponents"] = 4] = "WebComponents";
    TypeDescriptionPlatform[TypeDescriptionPlatform["XamarinForms"] = 5] = "XamarinForms";
    TypeDescriptionPlatform[TypeDescriptionPlatform["XamarinAndroid"] = 6] = "XamarinAndroid";
    TypeDescriptionPlatform[TypeDescriptionPlatform["XamariniOS"] = 7] = "XamariniOS";
    TypeDescriptionPlatform[TypeDescriptionPlatform["WindowsForms"] = 8] = "WindowsForms";
    TypeDescriptionPlatform[TypeDescriptionPlatform["UWP"] = 9] = "UWP";
    TypeDescriptionPlatform[TypeDescriptionPlatform["WinUI"] = 10] = "WinUI";
    TypeDescriptionPlatform[TypeDescriptionPlatform["Unknown"] = 11] = "Unknown";
    TypeDescriptionPlatform[TypeDescriptionPlatform["Blazor"] = 12] = "Blazor";
    return TypeDescriptionPlatform;
})({});
/**
 * @hidden
 */
export let TypeDescriptionPlatform_$type = /*@__PURE__*/ markEnum('TypeDescriptionPlatform', 'WPF,0|JQuery,1|React,2|Angular,3|WebComponents,4|XamarinForms,5|XamarinAndroid,6|XamariniOS,7|WindowsForms,8|UWP,9|WinUI,10|Unknown,11|Blazor,12');
