// igniteui-react-core/esm2015/lib/ValueBrushScaleDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BrushScaleDescription } from "./BrushScaleDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ValueBrushScaleDescription = /*@__PURE__*/ (() => {
    class ValueBrushScaleDescription extends BrushScaleDescription {
        get_type() {
            return "ValueBrushScale";
        }
        constructor() {
            super();
            this.p = 0;
            this.o = 0;
            this.m = false;
            this.s = 0;
        }
        get minimumValue() {
            return this.p;
        }
        set minimumValue(a) {
            this.p = a;
            this.g("MinimumValue");
        }
        get maximumValue() {
            return this.o;
        }
        set maximumValue(a) {
            this.o = a;
            this.g("MaximumValue");
        }
        get isLogarithmic() {
            return this.m;
        }
        set isLogarithmic(a) {
            this.m = a;
            this.g("IsLogarithmic");
        }
        get logarithmBase() {
            return this.s;
        }
        set logarithmBase(a) {
            this.s = a;
            this.g("LogarithmBase");
        }
    }
    ValueBrushScaleDescription.$t = /*@__PURE__*/ markType(ValueBrushScaleDescription, 'ValueBrushScaleDescription', BrushScaleDescription.$);
    return ValueBrushScaleDescription;
})();
