// igniteui-react-core/esm2015/lib/VersionQueueActionTimeout.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
/**
 * @hidden
 */
export let VersionQueueActionTimeout = /*@__PURE__*/ (() => {
    class VersionQueueActionTimeout extends Base {
        constructor(a, b, c) {
            super();
            this.d = 0;
            this.a = null;
            this.c = 0;
            this.a = a;
            this.d = b;
            this.c = c;
        }
        get f() {
            return this.d;
        }
        get b() {
            return this.a;
        }
        get e() {
            return this.c;
        }
    }
    VersionQueueActionTimeout.$t = /*@__PURE__*/ markType(VersionQueueActionTimeout, 'VersionQueueActionTimeout');
    return VersionQueueActionTimeout;
})();
