// igniteui-react-core/esm2015/lib/VerticalAnchoredCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnchoredCategorySeriesDescription } from "./AnchoredCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let VerticalAnchoredCategorySeriesDescription = /*@__PURE__*/ (() => {
    class VerticalAnchoredCategorySeriesDescription extends AnchoredCategorySeriesDescription {
        get_type() {
            return "VerticalAnchoredCategorySeries";
        }
        constructor() {
            super();
            this.gv = null;
            this.gw = null;
        }
        get xAxisRef() {
            return this.gv;
        }
        set xAxisRef(a) {
            this.gv = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.gw;
        }
        set yAxisRef(a) {
            this.gw = a;
            this.g("YAxisRef");
        }
    }
    VerticalAnchoredCategorySeriesDescription.$t = /*@__PURE__*/ markType(VerticalAnchoredCategorySeriesDescription, 'VerticalAnchoredCategorySeriesDescription', AnchoredCategorySeriesDescription.$);
    return VerticalAnchoredCategorySeriesDescription;
})();
