// igniteui-react-core/esm2015/lib/VirtualDataSourceDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseDataSourceDescription } from "./BaseDataSourceDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let VirtualDataSourceDescription = /*@__PURE__*/ (() => {
    class VirtualDataSourceDescription extends BaseDataSourceDescription {
        get_type() {
            return "VirtualDataSource";
        }
        constructor() {
            super();
            this.a3 = false;
            this.a4 = false;
            this.ba = 0;
            this.a9 = 0;
            this.a8 = 0;
            this.be = null;
            this.a2 = false;
        }
        get isSectionCollapsable() {
            return this.a3;
        }
        set isSectionCollapsable(a) {
            this.a3 = a;
            this.g("IsSectionCollapsable");
        }
        get isSectionExpandedDefault() {
            return this.a4;
        }
        set isSectionExpandedDefault(a) {
            this.a4 = a;
            this.g("IsSectionExpandedDefault");
        }
        get pageSizeRequested() {
            return this.ba;
        }
        set pageSizeRequested(a) {
            this.ba = a;
            this.g("PageSizeRequested");
        }
        get maxCachedPages() {
            return this.a9;
        }
        set maxCachedPages(a) {
            this.a9 = a;
            this.g("MaxCachedPages");
        }
        get actualPageSize() {
            return this.a8;
        }
        set actualPageSize(a) {
            this.a8 = a;
            this.g("ActualPageSize");
        }
        get concurrencyTag() {
            return this.be;
        }
        set concurrencyTag(a) {
            this.be = a;
            this.g("ConcurrencyTag");
        }
        get isBatchingEnabled() {
            return this.a2;
        }
        set isBatchingEnabled(a) {
            this.a2 = a;
            this.g("IsBatchingEnabled");
        }
    }
    VirtualDataSourceDescription.$t = /*@__PURE__*/ markType(VirtualDataSourceDescription, 'VirtualDataSourceDescription', BaseDataSourceDescription.$);
    return VirtualDataSourceDescription;
})();
