// igniteui-react-core/esm2015/lib/WebActiveStepChangingArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebActiveStepChangingArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebActiveStepChangingArgsDetailDescription extends Description {
        get_type() {
            return "WebActiveStepChangingArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = 0;
            this.i = 0;
        }
        get oldIndex() {
            return this.j;
        }
        set oldIndex(a) {
            this.j = a;
            this.g("OldIndex");
        }
        get newIndex() {
            return this.i;
        }
        set newIndex(a) {
            this.i = a;
            this.g("NewIndex");
        }
    }
    WebActiveStepChangingArgsDetailDescription.$t = /*@__PURE__*/ markType(WebActiveStepChangingArgsDetailDescription, 'WebActiveStepChangingArgsDetailDescription', Description.$);
    WebActiveStepChangingArgsDetailDescription.__marshalByValue = true;
    WebActiveStepChangingArgsDetailDescription.__marshalByValueAlias = "ActiveStepChangingArgsDetail";
    return WebActiveStepChangingArgsDetailDescription;
})();
