// igniteui-react-core/esm2015/lib/WebActivePaneEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebActivePaneEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebActivePaneEventArgsDetailDescription extends Description {
        get_type() {
            return "WebActivePaneEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.i = null;
        }
        get newPane() {
            return this.h;
        }
        set newPane(a) {
            this.h = a;
            this.g("NewPane");
        }
        get oldPane() {
            return this.i;
        }
        set oldPane(a) {
            this.i = a;
            this.g("OldPane");
        }
    }
    WebActivePaneEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebActivePaneEventArgsDetailDescription, 'WebActivePaneEventArgsDetailDescription', Description.$);
    WebActivePaneEventArgsDetailDescription.__marshalByValue = true;
    WebActivePaneEventArgsDetailDescription.__marshalByValueAlias = "ActivePaneEventArgsDetail";
    return WebActivePaneEventArgsDetailDescription;
})();
