// igniteui-react-core/esm2015/lib/WebActiveNodeChangeEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebActiveNodeChangeEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebActiveNodeChangeEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebActiveNodeChangeEventArgsDetail";
        }
        constructor() {
            super();
            this.p = 0;
            this.n = 0;
            this.o = 0;
            this.u = null;
        }
        get row() {
            return this.p;
        }
        set row(a) {
            this.p = a;
            this.g("Row");
        }
        get column() {
            return this.n;
        }
        set column(a) {
            this.n = a;
            this.g("Column");
        }
        get level() {
            return this.o;
        }
        set level(a) {
            this.o = a;
            this.g("Level");
        }
        get tag() {
            return this.u;
        }
        set tag(a) {
            this.u = a;
            this.g("Tag");
        }
    }
    WebActiveNodeChangeEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebActiveNodeChangeEventArgsDetailDescription, 'WebActiveNodeChangeEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebActiveNodeChangeEventArgsDetailDescription.__marshalByValue1 = true;
    WebActiveNodeChangeEventArgsDetailDescription.__marshalByValueAlias1 = "ActiveNodeChangeEventArgsDetail";
    return WebActiveNodeChangeEventArgsDetailDescription;
})();
