// igniteui-react-core/esm2015/lib/WebActionStripDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebActionStripDescription = /*@__PURE__*/ (() => {
    class WebActionStripDescription extends Description {
        get_type() {
            return "WebActionStrip";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.l = false;
            this.j = null;
            this.n = null;
        }
        get actionButtons() {
            return this.h;
        }
        set actionButtons(a) {
            this.h = a;
            this.g("ActionButtons");
        }
        get hidden() {
            return this.l;
        }
        set hidden(a) {
            this.l = a;
            this.g("Hidden");
        }
        get resourceStrings() {
            return this.j;
        }
        set resourceStrings(a) {
            this.j = a;
            this.g("ResourceStrings");
        }
        get name() {
            return this.n;
        }
        set name(a) {
            this.n = a;
            this.g("Name");
        }
    }
    WebActionStripDescription.$t = /*@__PURE__*/ markType(WebActionStripDescription, 'WebActionStripDescription', Description.$);
    return WebActionStripDescription;
})();
