// igniteui-react-core/esm2015/lib/WebAvatarDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebAvatarDescription = /*@__PURE__*/ (() => {
    class WebAvatarDescription extends Description {
        get_type() {
            return "WebAvatar";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = null;
            this.h = null;
            this.i = null;
            this.j = null;
        }
        get src() {
            return this.k;
        }
        set src(a) {
            this.k = a;
            this.g("Src");
        }
        get alt() {
            return this.h;
        }
        set alt(a) {
            this.h = a;
            this.g("Alt");
        }
        get initials() {
            return this.i;
        }
        set initials(a) {
            this.i = a;
            this.g("Initials");
        }
        get shape() {
            return this.j;
        }
        set shape(a) {
            this.j = a;
            this.g("Shape");
        }
    }
    WebAvatarDescription.$t = /*@__PURE__*/ markType(WebAvatarDescription, 'WebAvatarDescription', Description.$);
    return WebAvatarDescription;
})();
