// igniteui-react-core/esm2015/lib/WebBaseComboBoxLikeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebBaseComboBoxLikeDescription = /*@__PURE__*/ (() => {
    class WebBaseComboBoxLikeDescription extends Description {
        get_type() {
            return "WebBaseComboBoxLike";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.i = false;
            this.h = false;
            this.j = false;
        }
        get emitEvent() {
            return this.n;
        }
        set emitEvent(a) {
            this.n = a;
            this.g("EmitEvent");
        }
        get keepOpenOnSelect() {
            return this.i;
        }
        set keepOpenOnSelect(a) {
            this.i = a;
            this.g("KeepOpenOnSelect");
        }
        get keepOpenOnOutsideClick() {
            return this.h;
        }
        set keepOpenOnOutsideClick(a) {
            this.h = a;
            this.g("KeepOpenOnOutsideClick");
        }
        get open() {
            return this.j;
        }
        set open(a) {
            this.j = a;
            this.g("Open");
        }
    }
    WebBaseComboBoxLikeDescription.$t = /*@__PURE__*/ markType(WebBaseComboBoxLikeDescription, 'WebBaseComboBoxLikeDescription', Description.$);
    return WebBaseComboBoxLikeDescription;
})();
