// igniteui-react-core/esm2015/lib/WebButtonGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebButtonGroupDescription = /*@__PURE__*/ (() => {
    class WebButtonGroupDescription extends Description {
        get_type() {
            return "WebButtonGroup";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = false;
            this.l = null;
            this.o = null;
            this.h = null;
            this.n = null;
            this.m = null;
        }
        get disabled() {
            return this.j;
        }
        set disabled(a) {
            this.j = a;
            this.g("Disabled");
        }
        get alignment() {
            return this.l;
        }
        set alignment(a) {
            this.l = a;
            this.g("Alignment");
        }
        get selection() {
            return this.o;
        }
        set selection(a) {
            this.o = a;
            this.g("Selection");
        }
        get selectedItems() {
            return this.h;
        }
        set selectedItems(a) {
            this.h = a;
            this.g("SelectedItems");
        }
        get selectRef() {
            return this.n;
        }
        set selectRef(a) {
            this.n = a;
            this.g("SelectRef");
        }
        get deselectRef() {
            return this.m;
        }
        set deselectRef(a) {
            this.m = a;
            this.g("DeselectRef");
        }
    }
    WebButtonGroupDescription.$t = /*@__PURE__*/ markType(WebButtonGroupDescription, 'WebButtonGroupDescription', Description.$);
    return WebButtonGroupDescription;
})();
