// igniteui-react-core/esm2015/lib/WebBannerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebBannerDescription = /*@__PURE__*/ (() => {
    class WebBannerDescription extends Description {
        get_type() {
            return "WebBanner";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = false;
            this.k = null;
            this.j = null;
        }
        get open() {
            return this.h;
        }
        set open(a) {
            this.h = a;
            this.g("Open");
        }
        get closingRef() {
            return this.k;
        }
        set closingRef(a) {
            this.k = a;
            this.g("ClosingRef");
        }
        get closedRef() {
            return this.j;
        }
        set closedRef(a) {
            this.j = a;
            this.g("ClosedRef");
        }
    }
    WebBannerDescription.$t = /*@__PURE__*/ markType(WebBannerDescription, 'WebBannerDescription', Description.$);
    return WebBannerDescription;
})();
