// igniteui-react-core/esm2015/lib/WebBadgeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebBadgeDescription = /*@__PURE__*/ (() => {
    class WebBadgeDescription extends Description {
        get_type() {
            return "WebBadge";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = null;
            this.h = false;
            this.j = null;
        }
        get variant() {
            return this.k;
        }
        set variant(a) {
            this.k = a;
            this.g("Variant");
        }
        get outlined() {
            return this.h;
        }
        set outlined(a) {
            this.h = a;
            this.g("Outlined");
        }
        get shape() {
            return this.j;
        }
        set shape(a) {
            this.j = a;
            this.g("Shape");
        }
    }
    WebBadgeDescription.$t = /*@__PURE__*/ markType(WebBadgeDescription, 'WebBadgeDescription', Description.$);
    return WebBadgeDescription;
})();
