// igniteui-react-core/esm2015/lib/WebBaseSearchInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebBaseSearchInfoDescription = /*@__PURE__*/ (() => {
    class WebBaseSearchInfoDescription extends Description {
        get_type() {
            return "WebBaseSearchInfo";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.o = null;
            this.h = false;
            this.i = false;
            this.l = 0;
            this.n = null;
        }
        get searchText() {
            return this.o;
        }
        set searchText(a) {
            this.o = a;
            this.g("SearchText");
        }
        get caseSensitive() {
            return this.h;
        }
        set caseSensitive(a) {
            this.h = a;
            this.g("CaseSensitive");
        }
        get exactMatch() {
            return this.i;
        }
        set exactMatch(a) {
            this.i = a;
            this.g("ExactMatch");
        }
        get matchCount() {
            return this.l;
        }
        set matchCount(a) {
            this.l = a;
            this.g("MatchCount");
        }
        get content() {
            return this.n;
        }
        set content(a) {
            this.n = a;
            this.g("Content");
        }
    }
    WebBaseSearchInfoDescription.$t = /*@__PURE__*/ markType(WebBaseSearchInfoDescription, 'WebBaseSearchInfoDescription', Description.$);
    return WebBaseSearchInfoDescription;
})();
