// igniteui-react-core/esm2015/lib/WebBaseAlertLikeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebBaseAlertLikeDescription = /*@__PURE__*/ (() => {
    class WebBaseAlertLikeDescription extends Description {
        get_type() {
            return "WebBaseAlertLike";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = false;
            this.l = 0;
            this.h = false;
            this.n = null;
        }
        get open() {
            return this.i;
        }
        set open(a) {
            this.i = a;
            this.g("Open");
        }
        get displayTime() {
            return this.l;
        }
        set displayTime(a) {
            this.l = a;
            this.g("DisplayTime");
        }
        get keepOpen() {
            return this.h;
        }
        set keepOpen(a) {
            this.h = a;
            this.g("KeepOpen");
        }
        get position() {
            return this.n;
        }
        set position(a) {
            this.n = a;
            this.g("Position");
        }
    }
    WebBaseAlertLikeDescription.$t = /*@__PURE__*/ markType(WebBaseAlertLikeDescription, 'WebBaseAlertLikeDescription', Description.$);
    return WebBaseAlertLikeDescription;
})();
