// igniteui-react-core/esm2015/lib/WebBaseToolbarDirectiveDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebGridToolbarBaseActionDescription } from "./WebGridToolbarBaseActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebBaseToolbarDirectiveDescription = /*@__PURE__*/ (() => {
    class WebBaseToolbarDirectiveDescription extends WebGridToolbarBaseActionDescription {
        get_type() {
            return "WebBaseToolbarDirective";
        }
        constructor() {
            super();
            this.o = null;
            this.u = null;
            this.t = null;
            this.k = null;
            this.q = null;
            this.s = null;
            this.r = null;
            this.n = null;
            this.m = null;
            this.p = null;
        }
        get columnListHeight() {
            return this.o;
        }
        set columnListHeight(a) {
            this.o = a;
            this.g("ColumnListHeight");
        }
        get title() {
            return this.u;
        }
        set title(a) {
            this.u = a;
            this.g("Title");
        }
        get prompt() {
            return this.t;
        }
        set prompt(a) {
            this.t = a;
            this.g("Prompt");
        }
        get overlaySettings() {
            return this.k;
        }
        set overlaySettings(a) {
            this.k = a;
            this.g("OverlaySettings");
        }
        get name() {
            return this.q;
        }
        set name(a) {
            this.q = a;
            this.g("Name");
        }
        get openingRef() {
            return this.s;
        }
        set openingRef(a) {
            this.s = a;
            this.g("OpeningRef");
        }
        get openedRef() {
            return this.r;
        }
        set openedRef(a) {
            this.r = a;
            this.g("OpenedRef");
        }
        get closingRef() {
            return this.n;
        }
        set closingRef(a) {
            this.n = a;
            this.g("ClosingRef");
        }
        get closedRef() {
            return this.m;
        }
        set closedRef(a) {
            this.m = a;
            this.g("ClosedRef");
        }
        get columnToggleRef() {
            return this.p;
        }
        set columnToggleRef(a) {
            this.p = a;
            this.g("ColumnToggleRef");
        }
    }
    WebBaseToolbarDirectiveDescription.$t = /*@__PURE__*/ markType(WebBaseToolbarDirectiveDescription, 'WebBaseToolbarDirectiveDescription', WebGridToolbarBaseActionDescription.$);
    return WebBaseToolbarDirectiveDescription;
})();
