// igniteui-react-core/esm2015/lib/WebButtonBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebButtonBaseDescription = /*@__PURE__*/ (() => {
    class WebButtonBaseDescription extends Description {
        get_type() {
            return "WebButtonBase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.o = null;
            this.m = null;
            this.q = null;
            this.p = null;
            this.h = false;
            this.k = null;
            this.n = null;
            this.j = null;
        }
        get displayType() {
            return this.l;
        }
        set displayType(a) {
            this.l = a;
            this.g("DisplayType");
        }
        get href() {
            return this.o;
        }
        set href(a) {
            this.o = a;
            this.g("Href");
        }
        get download() {
            return this.m;
        }
        set download(a) {
            this.m = a;
            this.g("Download");
        }
        get target() {
            return this.q;
        }
        set target(a) {
            this.q = a;
            this.g("Target");
        }
        get rel() {
            return this.p;
        }
        set rel(a) {
            this.p = a;
            this.g("Rel");
        }
        get disabled() {
            return this.h;
        }
        set disabled(a) {
            this.h = a;
            this.g("Disabled");
        }
        get clickedRef() {
            return this.k;
        }
        set clickedRef(a) {
            this.k = a;
            this.g("ClickedRef");
        }
        get focusRef() {
            return this.n;
        }
        set focusRef(a) {
            this.n = a;
            this.g("FocusRef");
        }
        get blurRef() {
            return this.j;
        }
        set blurRef(a) {
            this.j = a;
            this.g("BlurRef");
        }
    }
    WebButtonBaseDescription.$t = /*@__PURE__*/ markType(WebButtonBaseDescription, 'WebButtonBaseDescription', Description.$);
    return WebButtonBaseDescription;
})();
