// igniteui-react-core/esm2015/lib/WebComboDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebComboDescription = /*@__PURE__*/ (() => {
    class WebComboDescription extends Description {
        get_type() {
            return "WebCombo";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.al = null;
            this.s = false;
            this.u = false;
            this.l = false;
            this.m = false;
            this.as = null;
            this.aw = null;
            this.ax = null;
            this.r = false;
            this.ay = null;
            this.am = null;
            this.ap = null;
            this.aq = null;
            this.j = null;
            this.n = false;
            this.p = false;
            this.h = null;
            this.o = false;
            this.t = false;
            this.af = null;
            this.at = null;
            this.q = false;
            this.ar = null;
            this.ao = null;
            this.ai = null;
            this.an = null;
            this.ah = null;
            this.av = null;
            this.au = null;
            this.ak = null;
            this.aj = null;
        }
        get dataRef() {
            return this.al;
        }
        set dataRef(a) {
            this.al = a;
            this.g("DataRef");
        }
        get outlined() {
            return this.s;
        }
        set outlined(a) {
            this.s = a;
            this.g("Outlined");
        }
        get singleSelect() {
            return this.u;
        }
        set singleSelect(a) {
            this.u = a;
            this.g("SingleSelect");
        }
        get autofocus() {
            return this.l;
        }
        set autofocus(a) {
            this.l = a;
            this.g("Autofocus");
        }
        get autofocusList() {
            return this.m;
        }
        set autofocusList(a) {
            this.m = a;
            this.g("AutofocusList");
        }
        get label() {
            return this.as;
        }
        set label(a) {
            this.as = a;
            this.g("Label");
        }
        get placeholder() {
            return this.aw;
        }
        set placeholder(a) {
            this.aw = a;
            this.g("Placeholder");
        }
        get placeholderSearch() {
            return this.ax;
        }
        set placeholderSearch(a) {
            this.ax = a;
            this.g("PlaceholderSearch");
        }
        get open() {
            return this.r;
        }
        set open(a) {
            this.r = a;
            this.g("Open");
        }
        get valueKey() {
            return this.ay;
        }
        set valueKey(a) {
            this.ay = a;
            this.g("ValueKey");
        }
        get displayKey() {
            return this.am;
        }
        set displayKey(a) {
            this.am = a;
            this.g("DisplayKey");
        }
        get groupKey() {
            return this.ap;
        }
        set groupKey(a) {
            this.ap = a;
            this.g("GroupKey");
        }
        get groupSorting() {
            return this.aq;
        }
        set groupSorting(a) {
            this.aq = a;
            this.g("GroupSorting");
        }
        get filteringOptions() {
            return this.j;
        }
        set filteringOptions(a) {
            this.j = a;
            this.g("FilteringOptions");
        }
        get caseSensitiveIcon() {
            return this.n;
        }
        set caseSensitiveIcon(a) {
            this.n = a;
            this.g("CaseSensitiveIcon");
        }
        get disableFiltering() {
            return this.p;
        }
        set disableFiltering(a) {
            this.p = a;
            this.g("DisableFiltering");
        }
        get value() {
            return this.h;
        }
        set value(a) {
            this.h = a;
            this.g("Value");
        }
        get disabled() {
            return this.o;
        }
        set disabled(a) {
            this.o = a;
            this.g("Disabled");
        }
        get required() {
            return this.t;
        }
        set required(a) {
            this.t = a;
            this.g("Required");
        }
        get defaultValue() {
            return this.af;
        }
        set defaultValue(a) {
            this.af = a;
            this.g("DefaultValue");
        }
        get name() {
            return this.at;
        }
        set name(a) {
            this.at = a;
            this.g("Name");
        }
        get invalid() {
            return this.q;
        }
        set invalid(a) {
            this.q = a;
            this.g("Invalid");
        }
        get itemTemplateRef() {
            return this.ar;
        }
        set itemTemplateRef(a) {
            this.ar = a;
            this.g("ItemTemplateRef");
        }
        get groupHeaderTemplateRef() {
            return this.ao;
        }
        set groupHeaderTemplateRef(a) {
            this.ao = a;
            this.g("GroupHeaderTemplateRef");
        }
        get changeRef() {
            return this.ai;
        }
        set changeRef(a) {
            this.ai = a;
            this.g("ChangeRef");
        }
        get focusRef() {
            return this.an;
        }
        set focusRef(a) {
            this.an = a;
            this.g("FocusRef");
        }
        get blurRef() {
            return this.ah;
        }
        set blurRef(a) {
            this.ah = a;
            this.g("BlurRef");
        }
        get openingRef() {
            return this.av;
        }
        set openingRef(a) {
            this.av = a;
            this.g("OpeningRef");
        }
        get openedRef() {
            return this.au;
        }
        set openedRef(a) {
            this.au = a;
            this.g("OpenedRef");
        }
        get closingRef() {
            return this.ak;
        }
        set closingRef(a) {
            this.ak = a;
            this.g("ClosingRef");
        }
        get closedRef() {
            return this.aj;
        }
        set closedRef(a) {
            this.aj = a;
            this.g("ClosedRef");
        }
    }
    WebComboDescription.$t = /*@__PURE__*/ markType(WebComboDescription, 'WebComboDescription', Description.$);
    return WebComboDescription;
})();
