// igniteui-react-core/esm2015/lib/WebCircularGradientDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCircularGradientDescription = /*@__PURE__*/ (() => {
    class WebCircularGradientDescription extends Description {
        get_type() {
            return "WebCircularGradient";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = null;
            this.j = null;
            this.h = 0;
        }
        get offset() {
            return this.k;
        }
        set offset(a) {
            this.k = a;
            this.g("Offset");
        }
        get color() {
            return this.j;
        }
        set color(a) {
            this.j = a;
            this.g("Color");
        }
        get opacity() {
            return this.h;
        }
        set opacity(a) {
            this.h = a;
            this.g("Opacity");
        }
    }
    WebCircularGradientDescription.$t = /*@__PURE__*/ markType(WebCircularGradientDescription, 'WebCircularGradientDescription', Description.$);
    return WebCircularGradientDescription;
})();
