// igniteui-react-core/esm2015/lib/WebColumnSelectionEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnSelectionEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebColumnSelectionEventArgsDetailDescription extends Description {
        get_type() {
            return "WebColumnSelectionEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.i = null;
            this.h = null;
            this.k = null;
            this.q = false;
            this.s = null;
        }
        get oldSelection() {
            return this.j;
        }
        set oldSelection(a) {
            this.j = a;
            this.g("OldSelection");
        }
        get newSelection() {
            return this.i;
        }
        set newSelection(a) {
            this.i = a;
            this.g("NewSelection");
        }
        get added() {
            return this.h;
        }
        set added(a) {
            this.h = a;
            this.g("Added");
        }
        get removed() {
            return this.k;
        }
        set removed(a) {
            this.k = a;
            this.g("Removed");
        }
        get cancel() {
            return this.q;
        }
        set cancel(a) {
            this.q = a;
            this.g("Cancel");
        }
        get owner() {
            return this.s;
        }
        set owner(a) {
            this.s = a;
            this.g("Owner");
        }
    }
    WebColumnSelectionEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebColumnSelectionEventArgsDetailDescription, 'WebColumnSelectionEventArgsDetailDescription', Description.$);
    WebColumnSelectionEventArgsDetailDescription.__marshalByValue = true;
    WebColumnSelectionEventArgsDetailDescription.__marshalByValueAlias = "ColumnSelectionEventArgsDetail";
    return WebColumnSelectionEventArgsDetailDescription;
})();
