// igniteui-react-core/esm2015/lib/WebColumnResizeEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnResizeEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebColumnResizeEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebColumnResizeEventArgsDetail";
        }
        constructor() {
            super();
            this.m = null;
            this.r = null;
            this.q = null;
        }
        get column() {
            return this.m;
        }
        set column(a) {
            this.m = a;
            this.g("Column");
        }
        get prevWidth() {
            return this.r;
        }
        set prevWidth(a) {
            this.r = a;
            this.g("PrevWidth");
        }
        get newWidth() {
            return this.q;
        }
        set newWidth(a) {
            this.q = a;
            this.g("NewWidth");
        }
    }
    WebColumnResizeEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebColumnResizeEventArgsDetailDescription, 'WebColumnResizeEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebColumnResizeEventArgsDetailDescription.__marshalByValue1 = true;
    WebColumnResizeEventArgsDetailDescription.__marshalByValueAlias1 = "ColumnResizeEventArgsDetail";
    return WebColumnResizeEventArgsDetailDescription;
})();
