// igniteui-react-core/esm2015/lib/WebCheckboxBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCheckboxBaseDescription = /*@__PURE__*/ (() => {
    class WebCheckboxBaseDescription extends Description {
        get_type() {
            return "WebCheckboxBase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.w = null;
            this.h = false;
            this.u = null;
            this.j = false;
            this.l = false;
            this.i = false;
            this.v = null;
            this.k = false;
            this.s = null;
            this.t = null;
            this.r = null;
        }
        get value() {
            return this.w;
        }
        set value(a) {
            this.w = a;
            this.g("Value");
        }
        get checked() {
            return this.h;
        }
        set checked(a) {
            this.h = a;
            this.g("Checked");
        }
        get labelPosition() {
            return this.u;
        }
        set labelPosition(a) {
            this.u = a;
            this.g("LabelPosition");
        }
        get disabled() {
            return this.j;
        }
        set disabled(a) {
            this.j = a;
            this.g("Disabled");
        }
        get required() {
            return this.l;
        }
        set required(a) {
            this.l = a;
            this.g("Required");
        }
        get defaultChecked() {
            return this.i;
        }
        set defaultChecked(a) {
            this.i = a;
            this.g("DefaultChecked");
        }
        get name() {
            return this.v;
        }
        set name(a) {
            this.v = a;
            this.g("Name");
        }
        get invalid() {
            return this.k;
        }
        set invalid(a) {
            this.k = a;
            this.g("Invalid");
        }
        get changeRef() {
            return this.s;
        }
        set changeRef(a) {
            this.s = a;
            this.g("ChangeRef");
        }
        get focusRef() {
            return this.t;
        }
        set focusRef(a) {
            this.t = a;
            this.g("FocusRef");
        }
        get blurRef() {
            return this.r;
        }
        set blurRef(a) {
            this.r = a;
            this.g("BlurRef");
        }
    }
    WebCheckboxBaseDescription.$t = /*@__PURE__*/ markType(WebCheckboxBaseDescription, 'WebCheckboxBaseDescription', Description.$);
    return WebCheckboxBaseDescription;
})();
