// igniteui-react-core/esm2015/lib/WebCalendarResourceStringsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCalendarResourceStringsDescription = /*@__PURE__*/ (() => {
    class WebCalendarResourceStringsDescription extends Description {
        get_type() {
            return "WebCalendarResourceStrings";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.s = null;
            this.u = null;
            this.q = null;
            this.t = null;
            this.r = null;
            this.v = null;
            this.j = null;
            this.n = null;
            this.k = null;
            this.o = null;
            this.l = null;
            this.p = null;
            this.m = null;
            this.w = null;
        }
        get selectMonth() {
            return this.s;
        }
        set selectMonth(a) {
            this.s = a;
            this.g("SelectMonth");
        }
        get selectYear() {
            return this.u;
        }
        set selectYear(a) {
            this.u = a;
            this.g("SelectYear");
        }
        get selectDate() {
            return this.q;
        }
        set selectDate(a) {
            this.q = a;
            this.g("SelectDate");
        }
        get selectRange() {
            return this.t;
        }
        set selectRange(a) {
            this.t = a;
            this.g("SelectRange");
        }
        get selectedDate() {
            return this.r;
        }
        set selectedDate(a) {
            this.r = a;
            this.g("SelectedDate");
        }
        get startDate() {
            return this.v;
        }
        set startDate(a) {
            this.v = a;
            this.g("StartDate");
        }
        get endDate() {
            return this.j;
        }
        set endDate(a) {
            this.j = a;
            this.g("EndDate");
        }
        get previousMonth() {
            return this.n;
        }
        set previousMonth(a) {
            this.n = a;
            this.g("PreviousMonth");
        }
        get nextMonth() {
            return this.k;
        }
        set nextMonth(a) {
            this.k = a;
            this.g("NextMonth");
        }
        get previousYear() {
            return this.o;
        }
        set previousYear(a) {
            this.o = a;
            this.g("PreviousYear");
        }
        get nextYear() {
            return this.l;
        }
        set nextYear(a) {
            this.l = a;
            this.g("NextYear");
        }
        get previousYears() {
            return this.p;
        }
        set previousYears(a) {
            this.p = a;
            this.g("PreviousYears");
        }
        get nextYears() {
            return this.m;
        }
        set nextYears(a) {
            this.m = a;
            this.g("NextYears");
        }
        get weekLabel() {
            return this.w;
        }
        set weekLabel(a) {
            this.w = a;
            this.g("WeekLabel");
        }
    }
    WebCalendarResourceStringsDescription.$t = /*@__PURE__*/ markType(WebCalendarResourceStringsDescription, 'WebCalendarResourceStringsDescription', Description.$);
    WebCalendarResourceStringsDescription.__marshalByValue = true;
    WebCalendarResourceStringsDescription.__marshalByValueAlias = "CalendarResourceStrings";
    return WebCalendarResourceStringsDescription;
})();
