// igniteui-react-core/esm2015/lib/WebCellTemplateContextDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCellTemplateContextDescription = /*@__PURE__*/ (() => {
    class WebCellTemplateContextDescription extends Description {
        get_type() {
            return "WebCellTemplateContext";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.k = null;
            this.h = null;
        }
        get implicit() {
            return this.l;
        }
        set implicit(a) {
            this.l = a;
            this.g("Implicit");
        }
        get additionalTemplateContext() {
            return this.k;
        }
        set additionalTemplateContext(a) {
            this.k = a;
            this.g("AdditionalTemplateContext");
        }
        get cell() {
            return this.h;
        }
        set cell(a) {
            this.h = a;
            this.g("Cell");
        }
    }
    WebCellTemplateContextDescription.$t = /*@__PURE__*/ markType(WebCellTemplateContextDescription, 'WebCellTemplateContextDescription', Description.$);
    WebCellTemplateContextDescription.__marshalByValue = true;
    WebCellTemplateContextDescription.__marshalByValueAlias = "CellTemplateContext";
    return WebCellTemplateContextDescription;
})();
