// igniteui-react-core/esm2015/lib/WebColumnTemplateContextDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnTemplateContextDescription = /*@__PURE__*/ (() => {
    class WebColumnTemplateContextDescription extends Description {
        get_type() {
            return "WebColumnTemplateContext";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
        }
        get implicit() {
            return this.i;
        }
        set implicit(a) {
            this.i = a;
            this.g("Implicit");
        }
        get column() {
            return this.h;
        }
        set column(a) {
            this.h = a;
            this.g("Column");
        }
    }
    WebColumnTemplateContextDescription.$t = /*@__PURE__*/ markType(WebColumnTemplateContextDescription, 'WebColumnTemplateContextDescription', Description.$);
    WebColumnTemplateContextDescription.__marshalByValue = true;
    WebColumnTemplateContextDescription.__marshalByValueAlias = "ColumnTemplateContext";
    return WebColumnTemplateContextDescription;
})();
