// igniteui-react-core/esm2015/lib/WebColumnMovingEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnMovingEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebColumnMovingEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebColumnMovingEventArgsDetail";
        }
        constructor() {
            super();
            this.m = null;
            this.p = false;
        }
        get source() {
            return this.m;
        }
        set source(a) {
            this.m = a;
            this.g("Source");
        }
        get cancel() {
            return this.p;
        }
        set cancel(a) {
            this.p = a;
            this.g("Cancel");
        }
    }
    WebColumnMovingEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebColumnMovingEventArgsDetailDescription, 'WebColumnMovingEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebColumnMovingEventArgsDetailDescription.__marshalByValue1 = true;
    WebColumnMovingEventArgsDetailDescription.__marshalByValueAlias1 = "ColumnMovingEventArgsDetail";
    return WebColumnMovingEventArgsDetailDescription;
})();
