// igniteui-react-core/esm2015/lib/WebChipDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebChipDescription = /*@__PURE__*/ (() => {
    class WebChipDescription extends Description {
        get_type() {
            return "WebChip";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = false;
            this.i = false;
            this.j = false;
            this.k = false;
            this.r = null;
            this.p = null;
            this.q = null;
        }
        get disabled() {
            return this.h;
        }
        set disabled(a) {
            this.h = a;
            this.g("Disabled");
        }
        get removable() {
            return this.i;
        }
        set removable(a) {
            this.i = a;
            this.g("Removable");
        }
        get selectable() {
            return this.j;
        }
        set selectable(a) {
            this.j = a;
            this.g("Selectable");
        }
        get selected() {
            return this.k;
        }
        set selected(a) {
            this.k = a;
            this.g("Selected");
        }
        get variant() {
            return this.r;
        }
        set variant(a) {
            this.r = a;
            this.g("Variant");
        }
        get removeRef() {
            return this.p;
        }
        set removeRef(a) {
            this.p = a;
            this.g("RemoveRef");
        }
        get selectRef() {
            return this.q;
        }
        set selectRef(a) {
            this.q = a;
            this.g("SelectRef");
        }
    }
    WebChipDescription.$t = /*@__PURE__*/ markType(WebChipDescription, 'WebChipDescription', Description.$);
    return WebChipDescription;
})();
