// igniteui-react-core/esm2015/lib/WebCalendarBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCalendarBaseDescription = /*@__PURE__*/ (() => {
    class WebCalendarBaseDescription extends Description {
        get_type() {
            return "WebCalendarBase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.o = null;
            this.l = false;
            this.p = null;
            this.n = null;
            this.i = null;
            this.h = null;
        }
        get selection() {
            return this.o;
        }
        set selection(a) {
            this.o = a;
            this.g("Selection");
        }
        get showWeekNumbers() {
            return this.l;
        }
        set showWeekNumbers(a) {
            this.l = a;
            this.g("ShowWeekNumbers");
        }
        get weekStart() {
            return this.p;
        }
        set weekStart(a) {
            this.p = a;
            this.g("WeekStart");
        }
        get locale() {
            return this.n;
        }
        set locale(a) {
            this.n = a;
            this.g("Locale");
        }
        get specialDates() {
            return this.i;
        }
        set specialDates(a) {
            this.i = a;
            this.g("SpecialDates");
        }
        get disabledDates() {
            return this.h;
        }
        set disabledDates(a) {
            this.h = a;
            this.g("DisabledDates");
        }
    }
    WebCalendarBaseDescription.$t = /*@__PURE__*/ markType(WebCalendarBaseDescription, 'WebCalendarBaseDescription', Description.$);
    return WebCalendarBaseDescription;
})();
