// igniteui-react-core/esm2015/lib/WebCellPositionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCellPositionDescription = /*@__PURE__*/ (() => {
    class WebCellPositionDescription extends Description {
        get_type() {
            return "WebCellPosition";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = 0;
            this.i = 0;
        }
        get rowIndex() {
            return this.h;
        }
        set rowIndex(a) {
            this.h = a;
            this.g("RowIndex");
        }
        get visibleColumnIndex() {
            return this.i;
        }
        set visibleColumnIndex(a) {
            this.i = a;
            this.g("VisibleColumnIndex");
        }
    }
    WebCellPositionDescription.$t = /*@__PURE__*/ markType(WebCellPositionDescription, 'WebCellPositionDescription', Description.$);
    return WebCellPositionDescription;
})();
