// igniteui-react-core/esm2015/lib/WebCarouselDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCarouselDescription = /*@__PURE__*/ (() => {
    class WebCarouselDescription extends Description {
        get_type() {
            return "WebCarousel";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = false;
            this.i = false;
            this.k = false;
            this.j = false;
            this.l = false;
            this.x = null;
            this.w = null;
            this.ab = null;
            this.r = 0;
            this.s = 0;
            this.v = null;
            this.aa = null;
            this.z = null;
            this.y = null;
        }
        get disableLoop() {
            return this.h;
        }
        set disableLoop(a) {
            this.h = a;
            this.g("DisableLoop");
        }
        get disablePauseOnInteraction() {
            return this.i;
        }
        set disablePauseOnInteraction(a) {
            this.i = a;
            this.g("DisablePauseOnInteraction");
        }
        get hideNavigation() {
            return this.k;
        }
        set hideNavigation(a) {
            this.k = a;
            this.g("HideNavigation");
        }
        get hideIndicators() {
            return this.j;
        }
        set hideIndicators(a) {
            this.j = a;
            this.g("HideIndicators");
        }
        get vertical() {
            return this.l;
        }
        set vertical(a) {
            this.l = a;
            this.g("Vertical");
        }
        get indicatorsOrientation() {
            return this.x;
        }
        set indicatorsOrientation(a) {
            this.x = a;
            this.g("IndicatorsOrientation");
        }
        get indicatorsLabelFormat() {
            return this.w;
        }
        set indicatorsLabelFormat(a) {
            this.w = a;
            this.g("IndicatorsLabelFormat");
        }
        get slidesLabelFormat() {
            return this.ab;
        }
        set slidesLabelFormat(a) {
            this.ab = a;
            this.g("SlidesLabelFormat");
        }
        get interval() {
            return this.r;
        }
        set interval(a) {
            this.r = a;
            this.g("Interval");
        }
        get maximumIndicatorsCount() {
            return this.s;
        }
        set maximumIndicatorsCount(a) {
            this.s = a;
            this.g("MaximumIndicatorsCount");
        }
        get animationType() {
            return this.v;
        }
        set animationType(a) {
            this.v = a;
            this.g("AnimationType");
        }
        get slideChangedRef() {
            return this.aa;
        }
        set slideChangedRef(a) {
            this.aa = a;
            this.g("SlideChangedRef");
        }
        get playingRef() {
            return this.z;
        }
        set playingRef(a) {
            this.z = a;
            this.g("PlayingRef");
        }
        get pausedRef() {
            return this.y;
        }
        set pausedRef(a) {
            this.y = a;
            this.g("PausedRef");
        }
    }
    WebCarouselDescription.$t = /*@__PURE__*/ markType(WebCarouselDescription, 'WebCarouselDescription', Description.$);
    return WebCarouselDescription;
})();
