// igniteui-react-core/esm2015/lib/WebClipboardOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebClipboardOptionsDescription = /*@__PURE__*/ (() => {
    class WebClipboardOptionsDescription extends Description {
        get_type() {
            return "WebClipboardOptions";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = false;
            this.i = false;
            this.h = false;
            this.n = null;
        }
        get enabled() {
            return this.j;
        }
        set enabled(a) {
            this.j = a;
            this.g("Enabled");
        }
        get copyHeaders() {
            return this.i;
        }
        set copyHeaders(a) {
            this.i = a;
            this.g("CopyHeaders");
        }
        get copyFormatters() {
            return this.h;
        }
        set copyFormatters(a) {
            this.h = a;
            this.g("CopyFormatters");
        }
        get separator() {
            return this.n;
        }
        set separator(a) {
            this.n = a;
            this.g("Separator");
        }
    }
    WebClipboardOptionsDescription.$t = /*@__PURE__*/ markType(WebClipboardOptionsDescription, 'WebClipboardOptionsDescription', Description.$);
    return WebClipboardOptionsDescription;
})();
