// igniteui-react-core/esm2015/lib/WebChildGridRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebChildGridRowDescription = /*@__PURE__*/ (() => {
    class WebChildGridRowDescription extends Description {
        get_type() {
            return "WebChildGridRow";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.n = null;
            this.l = 0;
            this.j = false;
        }
        get layout() {
            return this.h;
        }
        set layout(a) {
            this.h = a;
            this.g("Layout");
        }
        get data() {
            return this.n;
        }
        set data(a) {
            this.n = a;
            this.g("Data");
        }
        get index() {
            return this.l;
        }
        set index(a) {
            this.l = a;
            this.g("Index");
        }
        get expanded() {
            return this.j;
        }
        set expanded(a) {
            this.j = a;
            this.g("Expanded");
        }
    }
    WebChildGridRowDescription.$t = /*@__PURE__*/ markType(WebChildGridRowDescription, 'WebChildGridRowDescription', Description.$);
    return WebChildGridRowDescription;
})();
