// igniteui-react-core/esm2015/lib/WebColumnExportingEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnExportingEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebColumnExportingEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebColumnExportingEventArgsDetail";
        }
        constructor() {
            super();
            this.x = null;
            this.w = null;
            this.t = 0;
            this.p = false;
            this.q = false;
            this.m = null;
        }
        get header() {
            return this.x;
        }
        set header(a) {
            this.x = a;
            this.g("Header");
        }
        get field() {
            return this.w;
        }
        set field(a) {
            this.w = a;
            this.g("Field");
        }
        get columnIndex() {
            return this.t;
        }
        set columnIndex(a) {
            this.t = a;
            this.g("ColumnIndex");
        }
        get cancel() {
            return this.p;
        }
        set cancel(a) {
            this.p = a;
            this.g("Cancel");
        }
        get skipFormatter() {
            return this.q;
        }
        set skipFormatter(a) {
            this.q = a;
            this.g("SkipFormatter");
        }
        get grid() {
            return this.m;
        }
        set grid(a) {
            this.m = a;
            this.g("Grid");
        }
    }
    WebColumnExportingEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebColumnExportingEventArgsDetailDescription, 'WebColumnExportingEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebColumnExportingEventArgsDetailDescription.__marshalByValue1 = true;
    WebColumnExportingEventArgsDetailDescription.__marshalByValueAlias1 = "ColumnExportingEventArgsDetail";
    return WebColumnExportingEventArgsDetailDescription;
})();
