// igniteui-react-core/esm2015/lib/WebColumnVisibilityChangedEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnVisibilityChangedEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebColumnVisibilityChangedEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebColumnVisibilityChangedEventArgsDetail";
        }
        constructor() {
            super();
            this.p = null;
            this.n = false;
        }
        get column() {
            return this.p;
        }
        set column(a) {
            this.p = a;
            this.g("Column");
        }
        get newValue() {
            return this.n;
        }
        set newValue(a) {
            this.n = a;
            this.g("NewValue");
        }
    }
    WebColumnVisibilityChangedEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebColumnVisibilityChangedEventArgsDetailDescription, 'WebColumnVisibilityChangedEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebColumnVisibilityChangedEventArgsDetailDescription.__marshalByValue1 = true;
    WebColumnVisibilityChangedEventArgsDetailDescription.__marshalByValueAlias1 = "ColumnVisibilityChangedEventArgsDetail";
    return WebColumnVisibilityChangedEventArgsDetailDescription;
})();
