// igniteui-react-core/esm2015/lib/WebDialogDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDialogDescription = /*@__PURE__*/ (() => {
    class WebDialogDescription extends Description {
        get_type() {
            return "WebDialog";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = false;
            this.h = false;
            this.i = false;
            this.k = false;
            this.s = null;
            this.r = null;
            this.q = null;
            this.p = null;
        }
        get keepOpenOnEscape() {
            return this.j;
        }
        set keepOpenOnEscape(a) {
            this.j = a;
            this.g("KeepOpenOnEscape");
        }
        get closeOnOutsideClick() {
            return this.h;
        }
        set closeOnOutsideClick(a) {
            this.h = a;
            this.g("CloseOnOutsideClick");
        }
        get hideDefaultAction() {
            return this.i;
        }
        set hideDefaultAction(a) {
            this.i = a;
            this.g("HideDefaultAction");
        }
        get open() {
            return this.k;
        }
        set open(a) {
            this.k = a;
            this.g("Open");
        }
        get title() {
            return this.s;
        }
        set title(a) {
            this.s = a;
            this.g("Title");
        }
        get returnValue() {
            return this.r;
        }
        set returnValue(a) {
            this.r = a;
            this.g("ReturnValue");
        }
        get closingRef() {
            return this.q;
        }
        set closingRef(a) {
            this.q = a;
            this.g("ClosingRef");
        }
        get closedRef() {
            return this.p;
        }
        set closedRef(a) {
            this.p = a;
            this.g("ClosedRef");
        }
    }
    WebDialogDescription.$t = /*@__PURE__*/ markType(WebDialogDescription, 'WebDialogDescription', Description.$);
    return WebDialogDescription;
})();
