// igniteui-react-core/esm2015/lib/WebDimensionsChangeDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDimensionsChangeDetailDescription = /*@__PURE__*/ (() => {
    class WebDimensionsChangeDetailDescription extends Description {
        get_type() {
            return "WebDimensionsChangeDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.l = null;
        }
        get dimensions() {
            return this.h;
        }
        set dimensions(a) {
            this.h = a;
            this.g("Dimensions");
        }
        get dimensionCollectionType() {
            return this.l;
        }
        set dimensionCollectionType(a) {
            this.l = a;
            this.g("DimensionCollectionType");
        }
    }
    WebDimensionsChangeDetailDescription.$t = /*@__PURE__*/ markType(WebDimensionsChangeDetailDescription, 'WebDimensionsChangeDetailDescription', Description.$);
    WebDimensionsChangeDetailDescription.__marshalByValue = true;
    WebDimensionsChangeDetailDescription.__marshalByValueAlias = "DimensionsChangeDetail";
    return WebDimensionsChangeDetailDescription;
})();
