// igniteui-react-core/esm2015/lib/WebDockManagerResourceStringsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDockManagerResourceStringsDescription = /*@__PURE__*/ (() => {
    class WebDockManagerResourceStringsDescription extends Description {
        get_type() {
            return "WebDockManagerResourceStrings";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.o = null;
            this.p = null;
            this.j = null;
            this.k = null;
            this.l = null;
            this.m = null;
            this.n = null;
            this.i = null;
        }
        get close() {
            return this.h;
        }
        set close(a) {
            this.h = a;
            this.g("Close");
        }
        get pin() {
            return this.o;
        }
        set pin(a) {
            this.o = a;
            this.g("Pin");
        }
        get unpin() {
            return this.p;
        }
        set unpin(a) {
            this.p = a;
            this.g("Unpin");
        }
        get maximize() {
            return this.j;
        }
        set maximize(a) {
            this.j = a;
            this.g("Maximize");
        }
        get minimize() {
            return this.k;
        }
        set minimize(a) {
            this.k = a;
            this.g("Minimize");
        }
        get moreOptions() {
            return this.l;
        }
        set moreOptions(a) {
            this.l = a;
            this.g("MoreOptions");
        }
        get moreTabs() {
            return this.m;
        }
        set moreTabs(a) {
            this.m = a;
            this.g("MoreTabs");
        }
        get panes() {
            return this.n;
        }
        set panes(a) {
            this.n = a;
            this.g("Panes");
        }
        get documents() {
            return this.i;
        }
        set documents(a) {
            this.i = a;
            this.g("Documents");
        }
    }
    WebDockManagerResourceStringsDescription.$t = /*@__PURE__*/ markType(WebDockManagerResourceStringsDescription, 'WebDockManagerResourceStringsDescription', Description.$);
    return WebDockManagerResourceStringsDescription;
})();
