// igniteui-react-core/esm2015/lib/WebDimensionsChangeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDimensionsChangeDescription = /*@__PURE__*/ (() => {
    class WebDimensionsChangeDescription extends Description {
        get_type() {
            return "WebDimensionsChange";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = null;
        }
        get dimensions() {
            return this.h;
        }
        set dimensions(a) {
            this.h = a;
            this.g("Dimensions");
        }
        get dimensionCollectionType() {
            return this.j;
        }
        set dimensionCollectionType(a) {
            this.j = a;
            this.g("DimensionCollectionType");
        }
    }
    WebDimensionsChangeDescription.$t = /*@__PURE__*/ markType(WebDimensionsChangeDescription, 'WebDimensionsChangeDescription', Description.$);
    return WebDimensionsChangeDescription;
})();
