// igniteui-react-core/esm2015/lib/WebDockManagerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDockManagerDescription = /*@__PURE__*/ (() => {
    class WebDockManagerDescription extends Description {
        get_type() {
            return "WebDockManager";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.l = null;
            this.p = null;
            this.h = null;
            this.v = false;
            this.w = false;
            this.ah = null;
            this.m = null;
            this.r = null;
            this.t = false;
            this.x = false;
            this.u = false;
            this.ai = null;
            this.y = false;
            this.af = null;
            this.aj = null;
            this.ag = null;
        }
        get layout() {
            return this.j;
        }
        set layout(a) {
            this.j = a;
            this.g("Layout");
        }
        get draggedPane() {
            return this.l;
        }
        set draggedPane(a) {
            this.l = a;
            this.g("DraggedPane");
        }
        get dropPosition() {
            return this.p;
        }
        set dropPosition(a) {
            this.p = a;
            this.g("DropPosition");
        }
        get activePane() {
            return this.h;
        }
        set activePane(a) {
            this.h = a;
            this.g("ActivePane");
        }
        get allowMaximize() {
            return this.v;
        }
        set allowMaximize(a) {
            this.v = a;
            this.g("AllowMaximize");
        }
        get containedInBoundaries() {
            return this.w;
        }
        set containedInBoundaries(a) {
            this.w = a;
            this.g("ContainedInBoundaries");
        }
        get showHeaderIconOnHover() {
            return this.ah;
        }
        set showHeaderIconOnHover(a) {
            this.ah = a;
            this.g("ShowHeaderIconOnHover");
        }
        get maximizedPane() {
            return this.m;
        }
        set maximizedPane(a) {
            this.m = a;
            this.g("MaximizedPane");
        }
        get resourceStrings() {
            return this.r;
        }
        set resourceStrings(a) {
            this.r = a;
            this.g("ResourceStrings");
        }
        get allowFloatingPanesResize() {
            return this.t;
        }
        set allowFloatingPanesResize(a) {
            this.t = a;
            this.g("AllowFloatingPanesResize");
        }
        get disableKeyboardNavigation() {
            return this.x;
        }
        set disableKeyboardNavigation(a) {
            this.x = a;
            this.g("DisableKeyboardNavigation");
        }
        get allowInnerDock() {
            return this.u;
        }
        set allowInnerDock(a) {
            this.u = a;
            this.g("AllowInnerDock");
        }
        get showPaneHeaders() {
            return this.ai;
        }
        set showPaneHeaders(a) {
            this.ai = a;
            this.g("ShowPaneHeaders");
        }
        get proximityDock() {
            return this.y;
        }
        set proximityDock(a) {
            this.y = a;
            this.g("ProximityDock");
        }
        get closeBehavior() {
            return this.af;
        }
        set closeBehavior(a) {
            this.af = a;
            this.g("CloseBehavior");
        }
        get unpinBehavior() {
            return this.aj;
        }
        set unpinBehavior(a) {
            this.aj = a;
            this.g("UnpinBehavior");
        }
        get layoutChangeRef() {
            return this.ag;
        }
        set layoutChangeRef(a) {
            this.ag = a;
            this.g("LayoutChangeRef");
        }
    }
    WebDockManagerDescription.$t = /*@__PURE__*/ markType(WebDockManagerDescription, 'WebDockManagerDescription', Description.$);
    return WebDockManagerDescription;
})();
