// igniteui-react-core/esm2015/lib/WebDocumentHostDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebDockManagerPaneDescription } from "./WebDockManagerPaneDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDocumentHostDescription = /*@__PURE__*/ (() => {
    class WebDocumentHostDescription extends WebDockManagerPaneDescription {
        get_type() {
            return "WebDocumentHost";
        }
        constructor() {
            super();
            this.s = null;
            this.t = null;
            this.m = null;
            this.p = 0;
        }
        get id() {
            return this.s;
        }
        set id(a) {
            this.s = a;
            this.g("Id");
        }
        get paneType() {
            return this.t;
        }
        set paneType(a) {
            this.t = a;
            this.g("PaneType");
        }
        get rootPane() {
            return this.m;
        }
        set rootPane(a) {
            this.m = a;
            this.g("RootPane");
        }
        get size() {
            return this.p;
        }
        set size(a) {
            this.p = a;
            this.g("Size");
        }
    }
    WebDocumentHostDescription.$t = /*@__PURE__*/ markType(WebDocumentHostDescription, 'WebDocumentHostDescription', WebDockManagerPaneDescription.$);
    WebDocumentHostDescription.__marshalByValue1 = true;
    WebDocumentHostDescription.__marshalByValueAlias1 = "DocumentHost";
    return WebDocumentHostDescription;
})();
