// igniteui-react-core/esm2015/lib/WebDockingIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDockingIndicatorDescription = /*@__PURE__*/ (() => {
    class WebDockingIndicatorDescription extends Description {
        get_type() {
            return "WebDockingIndicator";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = null;
            this.h = false;
            this.j = null;
        }
        get position() {
            return this.k;
        }
        set position(a) {
            this.k = a;
            this.g("Position");
        }
        get isRoot() {
            return this.h;
        }
        set isRoot(a) {
            this.h = a;
            this.g("IsRoot");
        }
        get direction() {
            return this.j;
        }
        set direction(a) {
            this.j = a;
            this.g("Direction");
        }
    }
    WebDockingIndicatorDescription.$t = /*@__PURE__*/ markType(WebDockingIndicatorDescription, 'WebDockingIndicatorDescription', Description.$);
    return WebDockingIndicatorDescription;
})();
