// igniteui-react-core/esm2015/lib/WebDateTimeInputDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebMaskInputBaseDescription } from "./WebMaskInputBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDateTimeInputDescription = /*@__PURE__*/ (() => {
    class WebDateTimeInputDescription extends WebMaskInputBaseDescription {
        get_type() {
            return "WebDateTimeInput";
        }
        constructor() {
            super();
            this.au = null;
            this.ao = new Date();
            this.an = new Date();
            this.am = new Date();
            this.at = null;
            this.ai = null;
            this.ak = false;
            this.av = null;
            this.as = null;
        }
        get inputFormat() {
            return this.au;
        }
        set inputFormat(a) {
            this.au = a;
            this.g("InputFormat");
        }
        get value() {
            return this.ao;
        }
        set value(a) {
            this.ao = a;
            this.g("Value");
        }
        get min() {
            return this.an;
        }
        set min(a) {
            this.an = a;
            this.g("Min");
        }
        get max() {
            return this.am;
        }
        set max(a) {
            this.am = a;
            this.g("Max");
        }
        get displayFormat() {
            return this.at;
        }
        set displayFormat(a) {
            this.at = a;
            this.g("DisplayFormat");
        }
        get spinDelta() {
            return this.ai;
        }
        set spinDelta(a) {
            this.ai = a;
            this.g("SpinDelta");
        }
        get spinLoop() {
            return this.ak;
        }
        set spinLoop(a) {
            this.ak = a;
            this.g("SpinLoop");
        }
        get locale() {
            return this.av;
        }
        set locale(a) {
            this.av = a;
            this.g("Locale");
        }
        get changeRef() {
            return this.as;
        }
        set changeRef(a) {
            this.as = a;
            this.g("ChangeRef");
        }
    }
    WebDateTimeInputDescription.$t = /*@__PURE__*/ markType(WebDateTimeInputDescription, 'WebDateTimeInputDescription', WebMaskInputBaseDescription.$);
    return WebDateTimeInputDescription;
})();
