// igniteui-react-core/esm2015/lib/WebDividerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDividerDescription = /*@__PURE__*/ (() => {
    class WebDividerDescription extends Description {
        get_type() {
            return "WebDivider";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = false;
            this.h = false;
            this.l = null;
        }
        get vertical() {
            return this.i;
        }
        set vertical(a) {
            this.i = a;
            this.g("Vertical");
        }
        get middle() {
            return this.h;
        }
        set middle(a) {
            this.h = a;
            this.g("Middle");
        }
        get lineType() {
            return this.l;
        }
        set lineType(a) {
            this.l = a;
            this.g("LineType");
        }
    }
    WebDividerDescription.$t = /*@__PURE__*/ markType(WebDividerDescription, 'WebDividerDescription', Description.$);
    return WebDividerDescription;
})();
