// igniteui-react-core/esm2015/lib/WebDockManagerLayoutDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDockManagerLayoutDescription = /*@__PURE__*/ (() => {
    class WebDockManagerLayoutDescription extends Description {
        get_type() {
            return "WebDockManagerLayout";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.j = null;
            this.h = null;
        }
        get name() {
            return this.n;
        }
        set name(a) {
            this.n = a;
            this.g("Name");
        }
        get rootPane() {
            return this.j;
        }
        set rootPane(a) {
            this.j = a;
            this.g("RootPane");
        }
        get floatingPanes() {
            return this.h;
        }
        set floatingPanes(a) {
            this.h = a;
            this.g("FloatingPanes");
        }
    }
    WebDockManagerLayoutDescription.$t = /*@__PURE__*/ markType(WebDockManagerLayoutDescription, 'WebDockManagerLayoutDescription', Description.$);
    WebDockManagerLayoutDescription.__marshalByValue = true;
    WebDockManagerLayoutDescription.__marshalByValueAlias = "DockManagerLayout";
    return WebDockManagerLayoutDescription;
})();
