// igniteui-react-core/esm2015/lib/WebDropdownDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseComboBoxLikeDescription } from "./WebBaseComboBoxLikeDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDropdownDescription = /*@__PURE__*/ (() => {
    class WebDropdownDescription extends WebBaseComboBoxLikeDescription {
        get_type() {
            return "WebDropdown";
        }
        constructor() {
            super();
            this.ab = null;
            this.ac = null;
            this.q = false;
            this.u = 0;
            this.r = false;
            this.aa = null;
            this.z = null;
            this.y = null;
            this.x = null;
            this.w = null;
        }
        get placement() {
            return this.ab;
        }
        set placement(a) {
            this.ab = a;
            this.g("Placement");
        }
        get scrollStrategy() {
            return this.ac;
        }
        set scrollStrategy(a) {
            this.ac = a;
            this.g("ScrollStrategy");
        }
        get flip() {
            return this.q;
        }
        set flip(a) {
            this.q = a;
            this.g("Flip");
        }
        get distance() {
            return this.u;
        }
        set distance(a) {
            this.u = a;
            this.g("Distance");
        }
        get sameWidth() {
            return this.r;
        }
        set sameWidth(a) {
            this.r = a;
            this.g("SameWidth");
        }
        get openingRef() {
            return this.aa;
        }
        set openingRef(a) {
            this.aa = a;
            this.g("OpeningRef");
        }
        get openedRef() {
            return this.z;
        }
        set openedRef(a) {
            this.z = a;
            this.g("OpenedRef");
        }
        get closingRef() {
            return this.y;
        }
        set closingRef(a) {
            this.y = a;
            this.g("ClosingRef");
        }
        get closedRef() {
            return this.x;
        }
        set closedRef(a) {
            this.x = a;
            this.g("ClosedRef");
        }
        get changeRef() {
            return this.w;
        }
        set changeRef(a) {
            this.w = a;
            this.g("ChangeRef");
        }
    }
    WebDropdownDescription.$t = /*@__PURE__*/ markType(WebDropdownDescription, 'WebDropdownDescription', WebBaseComboBoxLikeDescription.$);
    return WebDropdownDescription;
})();
