// igniteui-react-core/esm2015/lib/WebDockManagerPointDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebDockManagerPointDescription = /*@__PURE__*/ (() => {
    class WebDockManagerPointDescription extends Description {
        get_type() {
            return "WebDockManagerPoint";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = 0;
            this.j = 0;
        }
        get x() {
            return this.i;
        }
        set x(a) {
            this.i = a;
            this.g("X");
        }
        get y() {
            return this.j;
        }
        set y(a) {
            this.j = a;
            this.g("Y");
        }
    }
    WebDockManagerPointDescription.$t = /*@__PURE__*/ markType(WebDockManagerPointDescription, 'WebDockManagerPointDescription', Description.$);
    WebDockManagerPointDescription.__marshalByValue = true;
    WebDockManagerPointDescription.__marshalByValueAlias = "DockManagerPoint";
    return WebDockManagerPointDescription;
})();
