// igniteui-react-core/esm2015/lib/WebExporterEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebExporterEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebExporterEventArgsDetailDescription extends Description {
        get_type() {
            return "WebExporterEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = null;
            this.l = null;
            this.o = false;
        }
        get exporter() {
            return this.h;
        }
        set exporter(a) {
            this.h = a;
            this.g("Exporter");
        }
        get options() {
            return this.j;
        }
        set options(a) {
            this.j = a;
            this.g("Options");
        }
        get grid() {
            return this.l;
        }
        set grid(a) {
            this.l = a;
            this.g("Grid");
        }
        get cancel() {
            return this.o;
        }
        set cancel(a) {
            this.o = a;
            this.g("Cancel");
        }
    }
    WebExporterEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebExporterEventArgsDetailDescription, 'WebExporterEventArgsDetailDescription', Description.$);
    WebExporterEventArgsDetailDescription.__marshalByValue = true;
    WebExporterEventArgsDetailDescription.__marshalByValueAlias = "ExporterEventArgsDetail";
    return WebExporterEventArgsDetailDescription;
})();
