// igniteui-react-core/esm2015/lib/WebExporterOptionsBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebExporterOptionsBaseDescription = /*@__PURE__*/ (() => {
    class WebExporterOptionsBaseDescription extends Description {
        get_type() {
            return "WebExporterOptionsBase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = false;
            this.m = false;
            this.k = false;
            this.p = false;
            this.n = false;
            this.o = false;
            this.i = false;
            this.j = false;
            this.h = false;
            this.z = null;
        }
        get ignoreColumnsVisibility() {
            return this.l;
        }
        set ignoreColumnsVisibility(a) {
            this.l = a;
            this.g("IgnoreColumnsVisibility");
        }
        get ignoreFiltering() {
            return this.m;
        }
        set ignoreFiltering(a) {
            this.m = a;
            this.g("IgnoreFiltering");
        }
        get ignoreColumnsOrder() {
            return this.k;
        }
        set ignoreColumnsOrder(a) {
            this.k = a;
            this.g("IgnoreColumnsOrder");
        }
        get ignoreSorting() {
            return this.p;
        }
        set ignoreSorting(a) {
            this.p = a;
            this.g("IgnoreSorting");
        }
        get ignoreGrouping() {
            return this.n;
        }
        set ignoreGrouping(a) {
            this.n = a;
            this.g("IgnoreGrouping");
        }
        get ignoreMultiColumnHeaders() {
            return this.o;
        }
        set ignoreMultiColumnHeaders(a) {
            this.o = a;
            this.g("IgnoreMultiColumnHeaders");
        }
        get exportSummaries() {
            return this.i;
        }
        set exportSummaries(a) {
            this.i = a;
            this.g("ExportSummaries");
        }
        get freezeHeaders() {
            return this.j;
        }
        set freezeHeaders(a) {
            this.j = a;
            this.g("FreezeHeaders");
        }
        get alwaysExportHeaders() {
            return this.h;
        }
        set alwaysExportHeaders(a) {
            this.h = a;
            this.g("AlwaysExportHeaders");
        }
        get fileName() {
            return this.z;
        }
        set fileName(a) {
            this.z = a;
            this.g("FileName");
        }
    }
    WebExporterOptionsBaseDescription.$t = /*@__PURE__*/ markType(WebExporterOptionsBaseDescription, 'WebExporterOptionsBaseDescription', Description.$);
    return WebExporterOptionsBaseDescription;
})();
