// igniteui-react-core/esm2015/lib/WebExpansionPanelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebExpansionPanelDescription = /*@__PURE__*/ (() => {
    class WebExpansionPanelDescription extends Description {
        get_type() {
            return "WebExpansionPanel";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = false;
            this.h = false;
            this.n = null;
            this.p = null;
            this.o = null;
            this.m = null;
            this.l = null;
        }
        get open() {
            return this.i;
        }
        set open(a) {
            this.i = a;
            this.g("Open");
        }
        get disabled() {
            return this.h;
        }
        set disabled(a) {
            this.h = a;
            this.g("Disabled");
        }
        get indicatorPosition() {
            return this.n;
        }
        set indicatorPosition(a) {
            this.n = a;
            this.g("IndicatorPosition");
        }
        get openingRef() {
            return this.p;
        }
        set openingRef(a) {
            this.p = a;
            this.g("OpeningRef");
        }
        get openedRef() {
            return this.o;
        }
        set openedRef(a) {
            this.o = a;
            this.g("OpenedRef");
        }
        get closingRef() {
            return this.m;
        }
        set closingRef(a) {
            this.m = a;
            this.g("ClosingRef");
        }
        get closedRef() {
            return this.l;
        }
        set closedRef(a) {
            this.l = a;
            this.g("ClosedRef");
        }
    }
    WebExpansionPanelDescription.$t = /*@__PURE__*/ markType(WebExpansionPanelDescription, 'WebExpansionPanelDescription', Description.$);
    return WebExpansionPanelDescription;
})();
