// igniteui-react-core/esm2015/lib/WebFilteringOperationDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebFilteringOperationDescription = /*@__PURE__*/ (() => {
    class WebFilteringOperationDescription extends Description {
        get_type() {
            return "WebFilteringOperation";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.p = null;
            this.j = false;
            this.n = null;
            this.i = false;
            this.o = null;
        }
        get name() {
            return this.p;
        }
        set name(a) {
            this.p = a;
            this.g("Name");
        }
        get isUnary() {
            return this.j;
        }
        set isUnary(a) {
            this.j = a;
            this.g("IsUnary");
        }
        get iconName() {
            return this.n;
        }
        set iconName(a) {
            this.n = a;
            this.g("IconName");
        }
        get hidden() {
            return this.i;
        }
        set hidden(a) {
            this.i = a;
            this.g("Hidden");
        }
        get logicRef() {
            return this.o;
        }
        set logicRef(a) {
            this.o = a;
            this.g("LogicRef");
        }
    }
    WebFilteringOperationDescription.$t = /*@__PURE__*/ markType(WebFilteringOperationDescription, 'WebFilteringOperationDescription', Description.$);
    WebFilteringOperationDescription.__marshalByValue = true;
    WebFilteringOperationDescription.__marshalByValueAlias = "FilteringOperation";
    return WebFilteringOperationDescription;
})();
