// igniteui-react-core/esm2015/lib/WebFilteringEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebFilteringEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebFilteringEventArgsDetailDescription extends Description {
        get_type() {
            return "WebFilteringEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.m = null;
            this.k = false;
        }
        get filteringExpressions() {
            return this.h;
        }
        set filteringExpressions(a) {
            this.h = a;
            this.g("FilteringExpressions");
        }
        get owner() {
            return this.m;
        }
        set owner(a) {
            this.m = a;
            this.g("Owner");
        }
        get cancel() {
            return this.k;
        }
        set cancel(a) {
            this.k = a;
            this.g("Cancel");
        }
    }
    WebFilteringEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebFilteringEventArgsDetailDescription, 'WebFilteringEventArgsDetailDescription', Description.$);
    WebFilteringEventArgsDetailDescription.__marshalByValue = true;
    WebFilteringEventArgsDetailDescription.__marshalByValueAlias = "FilteringEventArgsDetail";
    return WebFilteringEventArgsDetailDescription;
})();
