// igniteui-react-core/esm2015/lib/WebFieldPipeArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebFieldPipeArgsDescription = /*@__PURE__*/ (() => {
    class WebFieldPipeArgsDescription extends Description {
        get_type() {
            return "WebFieldPipeArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.m = null;
            this.n = null;
            this.k = null;
            this.j = null;
            this.l = null;
            this.h = 0;
        }
        get format() {
            return this.m;
        }
        set format(a) {
            this.m = a;
            this.g("Format");
        }
        get timezone() {
            return this.n;
        }
        set timezone(a) {
            this.n = a;
            this.g("Timezone");
        }
        get digitsInfo() {
            return this.k;
        }
        set digitsInfo(a) {
            this.k = a;
            this.g("DigitsInfo");
        }
        get currencyCode() {
            return this.j;
        }
        set currencyCode(a) {
            this.j = a;
            this.g("CurrencyCode");
        }
        get display() {
            return this.l;
        }
        set display(a) {
            this.l = a;
            this.g("Display");
        }
        get weekStart() {
            return this.h;
        }
        set weekStart(a) {
            this.h = a;
            this.g("WeekStart");
        }
    }
    WebFieldPipeArgsDescription.$t = /*@__PURE__*/ markType(WebFieldPipeArgsDescription, 'WebFieldPipeArgsDescription', Description.$);
    return WebFieldPipeArgsDescription;
})();
