// igniteui-react-core/esm2015/lib/WebForOfStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDescription } from "./WebBaseEventArgsDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebForOfStateDescription = /*@__PURE__*/ (() => {
    class WebForOfStateDescription extends WebBaseEventArgsDescription {
        get_type() {
            return "WebForOfState";
        }
        constructor() {
            super();
            this.o = 0;
            this.n = 0;
        }
        get startIndex() {
            return this.o;
        }
        set startIndex(a) {
            this.o = a;
            this.g("StartIndex");
        }
        get chunkSize() {
            return this.n;
        }
        set chunkSize(a) {
            this.n = a;
            this.g("ChunkSize");
        }
    }
    WebForOfStateDescription.$t = /*@__PURE__*/ markType(WebForOfStateDescription, 'WebForOfStateDescription', WebBaseEventArgsDescription.$);
    WebForOfStateDescription.__marshalByValue1 = true;
    WebForOfStateDescription.__marshalByValueAlias1 = "ForOfState";
    return WebForOfStateDescription;
})();
