// igniteui-react-core/esm2015/lib/WebFloatPaneActionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebPaneDragActionDescription } from "./WebPaneDragActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebFloatPaneActionDescription = /*@__PURE__*/ (() => {
    class WebFloatPaneActionDescription extends WebPaneDragActionDescription {
        get_type() {
            return "WebFloatPaneAction";
        }
        constructor() {
            super();
            this.s = null;
            this.k = null;
            this.o = 0;
            this.n = 0;
        }
        get actionType() {
            return this.s;
        }
        set actionType(a) {
            this.s = a;
            this.g("ActionType");
        }
        get location() {
            return this.k;
        }
        set location(a) {
            this.k = a;
            this.g("Location");
        }
        get width() {
            return this.o;
        }
        set width(a) {
            this.o = a;
            this.g("Width");
        }
        get height() {
            return this.n;
        }
        set height(a) {
            this.n = a;
            this.g("Height");
        }
    }
    WebFloatPaneActionDescription.$t = /*@__PURE__*/ markType(WebFloatPaneActionDescription, 'WebFloatPaneActionDescription', WebPaneDragActionDescription.$);
    WebFloatPaneActionDescription.__marshalByValue1 = true;
    WebFloatPaneActionDescription.__marshalByValueAlias1 = "FloatPaneAction";
    return WebFloatPaneActionDescription;
})();
