// igniteui-react-core/esm2015/lib/WebFilteringExpressionsTreeDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebFilteringExpressionsTreeDetailDescription = /*@__PURE__*/ (() => {
    class WebFilteringExpressionsTreeDetailDescription extends Description {
        get_type() {
            return "WebFilteringExpressionsTreeDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.p = null;
            this.k = null;
            this.o = null;
            this.n = null;
        }
        get filteringOperands() {
            return this.h;
        }
        set filteringOperands(a) {
            this.h = a;
            this.g("FilteringOperands");
        }
        get treeType() {
            return this.p;
        }
        set treeType(a) {
            this.p = a;
            this.g("TreeType");
        }
        get owner() {
            return this.k;
        }
        set owner(a) {
            this.k = a;
            this.g("Owner");
        }
        get operator() {
            return this.o;
        }
        set operator(a) {
            this.o = a;
            this.g("Operator");
        }
        get fieldName() {
            return this.n;
        }
        set fieldName(a) {
            this.n = a;
            this.g("FieldName");
        }
    }
    WebFilteringExpressionsTreeDetailDescription.$t = /*@__PURE__*/ markType(WebFilteringExpressionsTreeDetailDescription, 'WebFilteringExpressionsTreeDetailDescription', Description.$);
    WebFilteringExpressionsTreeDetailDescription.__marshalByValue = true;
    WebFilteringExpressionsTreeDetailDescription.__marshalByValueAlias = "FilteringExpressionsTreeDetail";
    return WebFilteringExpressionsTreeDetailDescription;
})();
