// igniteui-react-core/esm2015/lib/WebFloatingPaneResizeMoveEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebFloatingPaneResizeEventArgsDetailDescription } from "./WebFloatingPaneResizeEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebFloatingPaneResizeMoveEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebFloatingPaneResizeMoveEventArgsDetailDescription extends WebFloatingPaneResizeEventArgsDetailDescription {
        get_type() {
            return "WebFloatingPaneResizeMoveEventArgsDetail";
        }
        constructor() {
            super();
            this.w = 0;
            this.u = 0;
            this.v = 0;
            this.t = 0;
            this.p = null;
            this.o = null;
        }
        get oldWidth() {
            return this.w;
        }
        set oldWidth(a) {
            this.w = a;
            this.g("OldWidth");
        }
        get newWidth() {
            return this.u;
        }
        set newWidth(a) {
            this.u = a;
            this.g("NewWidth");
        }
        get oldHeight() {
            return this.v;
        }
        set oldHeight(a) {
            this.v = a;
            this.g("OldHeight");
        }
        get newHeight() {
            return this.t;
        }
        set newHeight(a) {
            this.t = a;
            this.g("NewHeight");
        }
        get oldLocation() {
            return this.p;
        }
        set oldLocation(a) {
            this.p = a;
            this.g("OldLocation");
        }
        get newLocation() {
            return this.o;
        }
        set newLocation(a) {
            this.o = a;
            this.g("NewLocation");
        }
    }
    WebFloatingPaneResizeMoveEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebFloatingPaneResizeMoveEventArgsDetailDescription, 'WebFloatingPaneResizeMoveEventArgsDetailDescription', WebFloatingPaneResizeEventArgsDetailDescription.$);
    WebFloatingPaneResizeMoveEventArgsDetailDescription.__marshalByValue1 = true;
    WebFloatingPaneResizeMoveEventArgsDetailDescription.__marshalByValueAlias1 = "FloatingPaneResizeMoveEventArgsDetail";
    return WebFloatingPaneResizeMoveEventArgsDetailDescription;
})();
