// igniteui-react-core/esm2015/lib/WebForOfDataChangingEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebForOfDataChangingEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebForOfDataChangingEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebForOfDataChangingEventArgsDetail";
        }
        constructor() {
            super();
            this.p = 0;
            this.m = null;
        }
        get containerSize() {
            return this.p;
        }
        set containerSize(a) {
            this.p = a;
            this.g("ContainerSize");
        }
        get state() {
            return this.m;
        }
        set state(a) {
            this.m = a;
            this.g("State");
        }
    }
    WebForOfDataChangingEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebForOfDataChangingEventArgsDetailDescription, 'WebForOfDataChangingEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebForOfDataChangingEventArgsDetailDescription.__marshalByValue1 = true;
    WebForOfDataChangingEventArgsDetailDescription.__marshalByValueAlias1 = "ForOfDataChangingEventArgsDetail";
    return WebForOfDataChangingEventArgsDetailDescription;
})();
