// igniteui-react-core/esm2015/lib/WebGridContextMenuEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridContextMenuEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridContextMenuEventArgsDetailDescription extends Description {
        get_type() {
            return "WebGridContextMenuEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.m = null;
            this.j = null;
        }
        get cell() {
            return this.h;
        }
        set cell(a) {
            this.h = a;
            this.g("Cell");
        }
        get event() {
            return this.m;
        }
        set event(a) {
            this.m = a;
            this.g("Event");
        }
        get row() {
            return this.j;
        }
        set row(a) {
            this.j = a;
            this.g("Row");
        }
    }
    WebGridContextMenuEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridContextMenuEventArgsDetailDescription, 'WebGridContextMenuEventArgsDetailDescription', Description.$);
    WebGridContextMenuEventArgsDetailDescription.__marshalByValue = true;
    WebGridContextMenuEventArgsDetailDescription.__marshalByValueAlias = "GridContextMenuEventArgsDetail";
    return WebGridContextMenuEventArgsDetailDescription;
})();
