// igniteui-react-core/esm2015/lib/WebGridCellEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridCellEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridCellEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebGridCellEventArgsDetail";
        }
        constructor() {
            super();
            this.m = null;
            this.p = null;
        }
        get cell() {
            return this.m;
        }
        set cell(a) {
            this.m = a;
            this.g("Cell");
        }
        get event() {
            return this.p;
        }
        set event(a) {
            this.p = a;
            this.g("Event");
        }
    }
    WebGridCellEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridCellEventArgsDetailDescription, 'WebGridCellEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebGridCellEventArgsDetailDescription.__marshalByValue1 = true;
    WebGridCellEventArgsDetailDescription.__marshalByValueAlias1 = "GridCellEventArgsDetail";
    return WebGridCellEventArgsDetailDescription;
})();
