// igniteui-react-core/esm2015/lib/WebGridCellIdentifierDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridCellIdentifierDescription = /*@__PURE__*/ (() => {
    class WebGridCellIdentifierDescription extends Description {
        get_type() {
            return "WebGridCellIdentifier";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.i = 0;
            this.j = 0;
        }
        get rowID() {
            return this.n;
        }
        set rowID(a) {
            this.n = a;
            this.g("RowID");
        }
        get columnID() {
            return this.i;
        }
        set columnID(a) {
            this.i = a;
            this.g("ColumnID");
        }
        get rowIndex() {
            return this.j;
        }
        set rowIndex(a) {
            this.j = a;
            this.g("RowIndex");
        }
    }
    WebGridCellIdentifierDescription.$t = /*@__PURE__*/ markType(WebGridCellIdentifierDescription, 'WebGridCellIdentifierDescription', Description.$);
    WebGridCellIdentifierDescription.__marshalByValue = true;
    WebGridCellIdentifierDescription.__marshalByValueAlias = "GridCellIdentifier";
    return WebGridCellIdentifierDescription;
})();
