// igniteui-react-core/esm2015/lib/WebGridToolbarExportEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridToolbarExportEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridToolbarExportEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebGridToolbarExportEventArgsDetail";
        }
        constructor() {
            super();
            this.q = null;
            this.m = null;
            this.o = null;
            this.t = false;
        }
        get grid() {
            return this.q;
        }
        set grid(a) {
            this.q = a;
            this.g("Grid");
        }
        get exporter() {
            return this.m;
        }
        set exporter(a) {
            this.m = a;
            this.g("Exporter");
        }
        get options() {
            return this.o;
        }
        set options(a) {
            this.o = a;
            this.g("Options");
        }
        get cancel() {
            return this.t;
        }
        set cancel(a) {
            this.t = a;
            this.g("Cancel");
        }
    }
    WebGridToolbarExportEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridToolbarExportEventArgsDetailDescription, 'WebGridToolbarExportEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebGridToolbarExportEventArgsDetailDescription.__marshalByValue1 = true;
    WebGridToolbarExportEventArgsDetailDescription.__marshalByValueAlias1 = "GridToolbarExportEventArgsDetail";
    return WebGridToolbarExportEventArgsDetailDescription;
})();
