// igniteui-react-core/esm2015/lib/WebGridStateCollectionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridStateCollectionDescription = /*@__PURE__*/ (() => {
    class WebGridStateCollectionDescription extends Description {
        get_type() {
            return "WebGridStateCollection";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.k = null;
            this.h = null;
        }
        get id() {
            return this.n;
        }
        set id(a) {
            this.n = a;
            this.g("Id");
        }
        get parentRowID() {
            return this.k;
        }
        set parentRowID(a) {
            this.k = a;
            this.g("ParentRowID");
        }
        get state() {
            return this.h;
        }
        set state(a) {
            this.h = a;
            this.g("State");
        }
    }
    WebGridStateCollectionDescription.$t = /*@__PURE__*/ markType(WebGridStateCollectionDescription, 'WebGridStateCollectionDescription', Description.$);
    WebGridStateCollectionDescription.__marshalByValue = true;
    WebGridStateCollectionDescription.__marshalByValueAlias = "GridStateCollection";
    return WebGridStateCollectionDescription;
})();
