// igniteui-react-core/esm2015/lib/WebGridStateBaseDirectiveDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridStateBaseDirectiveDescription = /*@__PURE__*/ (() => {
    class WebGridStateBaseDirectiveDescription extends Description {
        get_type() {
            return "WebGridStateBaseDirective";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = null;
        }
        get options() {
            return this.h;
        }
        set options(a) {
            this.h = a;
            this.g("Options");
        }
        get name() {
            return this.j;
        }
        set name(a) {
            this.j = a;
            this.g("Name");
        }
    }
    WebGridStateBaseDirectiveDescription.$t = /*@__PURE__*/ markType(WebGridStateBaseDirectiveDescription, 'WebGridStateBaseDirectiveDescription', Description.$);
    return WebGridStateBaseDirectiveDescription;
})();
