// igniteui-react-core/esm2015/lib/WebGroupByResultDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGroupByResultDescription = /*@__PURE__*/ (() => {
    class WebGroupByResultDescription extends Description {
        get_type() {
            return "WebGroupByResult";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = null;
        }
        get dataRef() {
            return this.j;
        }
        set dataRef(a) {
            this.j = a;
            this.g("DataRef");
        }
        get metadata() {
            return this.h;
        }
        set metadata(a) {
            this.h = a;
            this.g("Metadata");
        }
    }
    WebGroupByResultDescription.$t = /*@__PURE__*/ markType(WebGroupByResultDescription, 'WebGroupByResultDescription', Description.$);
    return WebGroupByResultDescription;
})();
