// igniteui-react-core/esm2015/lib/WebGridActionsBaseDirectiveDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridActionsBaseDirectiveDescription = /*@__PURE__*/ (() => {
    class WebGridActionsBaseDirectiveDescription extends Description {
        get_type() {
            return "WebGridActionsBaseDirective";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = false;
            this.j = null;
        }
        get asMenuItems() {
            return this.h;
        }
        set asMenuItems(a) {
            this.h = a;
            this.g("AsMenuItems");
        }
        get name() {
            return this.j;
        }
        set name(a) {
            this.j = a;
            this.g("Name");
        }
    }
    WebGridActionsBaseDirectiveDescription.$t = /*@__PURE__*/ markType(WebGridActionsBaseDirectiveDescription, 'WebGridActionsBaseDirectiveDescription', Description.$);
    return WebGridActionsBaseDirectiveDescription;
})();
