// igniteui-react-core/esm2015/lib/WebGridScrollEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridScrollEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridScrollEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebGridScrollEventArgsDetail";
        }
        constructor() {
            super();
            this.s = null;
            this.p = null;
            this.n = 0;
        }
        get direction() {
            return this.s;
        }
        set direction(a) {
            this.s = a;
            this.g("Direction");
        }
        get event() {
            return this.p;
        }
        set event(a) {
            this.p = a;
            this.g("Event");
        }
        get scrollPosition() {
            return this.n;
        }
        set scrollPosition(a) {
            this.n = a;
            this.g("ScrollPosition");
        }
    }
    WebGridScrollEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridScrollEventArgsDetailDescription, 'WebGridScrollEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebGridScrollEventArgsDetailDescription.__marshalByValue1 = true;
    WebGridScrollEventArgsDetailDescription.__marshalByValueAlias1 = "GridScrollEventArgsDetail";
    return WebGridScrollEventArgsDetailDescription;
})();
