// igniteui-react-core/esm2015/lib/WebGridKeydownEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridKeydownEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridKeydownEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebGridKeydownEventArgsDetail";
        }
        constructor() {
            super();
            this.u = null;
            this.q = null;
            this.p = null;
            this.n = false;
        }
        get targetType() {
            return this.u;
        }
        set targetType(a) {
            this.u = a;
            this.g("TargetType");
        }
        get target() {
            return this.q;
        }
        set target(a) {
            this.q = a;
            this.g("Target");
        }
        get event() {
            return this.p;
        }
        set event(a) {
            this.p = a;
            this.g("Event");
        }
        get cancel() {
            return this.n;
        }
        set cancel(a) {
            this.n = a;
            this.g("Cancel");
        }
    }
    WebGridKeydownEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridKeydownEventArgsDetailDescription, 'WebGridKeydownEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebGridKeydownEventArgsDetailDescription.__marshalByValue1 = true;
    WebGridKeydownEventArgsDetailDescription.__marshalByValueAlias1 = "GridKeydownEventArgsDetail";
    return WebGridKeydownEventArgsDetailDescription;
})();
