// igniteui-react-core/esm2015/lib/WebGridEditDoneEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridEditDoneEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridEditDoneEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebGridEditDoneEventArgsDetail";
        }
        constructor() {
            super();
            this.ab = null;
            this.z = null;
            this.ac = null;
            this.q = null;
            this.aa = null;
            this.y = null;
            this.x = null;
            this.m = null;
            this.o = null;
            this.t = false;
            this.u = false;
        }
        get rowID() {
            return this.ab;
        }
        set rowID(a) {
            this.ab = a;
            this.g("RowID");
        }
        get primaryKey() {
            return this.z;
        }
        set primaryKey(a) {
            this.z = a;
            this.g("PrimaryKey");
        }
        get rowKey() {
            return this.ac;
        }
        set rowKey(a) {
            this.ac = a;
            this.g("RowKey");
        }
        get cellID() {
            return this.q;
        }
        set cellID(a) {
            this.q = a;
            this.g("CellID");
        }
        get rowData() {
            return this.aa;
        }
        set rowData(a) {
            this.aa = a;
            this.g("RowData");
        }
        get oldValue() {
            return this.y;
        }
        set oldValue(a) {
            this.y = a;
            this.g("OldValue");
        }
        get newValue() {
            return this.x;
        }
        set newValue(a) {
            this.x = a;
            this.g("NewValue");
        }
        get column() {
            return this.m;
        }
        set column(a) {
            this.m = a;
            this.g("Column");
        }
        get owner() {
            return this.o;
        }
        set owner(a) {
            this.o = a;
            this.g("Owner");
        }
        get isAddRow() {
            return this.t;
        }
        set isAddRow(a) {
            this.t = a;
            this.g("IsAddRow");
        }
        get valid() {
            return this.u;
        }
        set valid(a) {
            this.u = a;
            this.g("Valid");
        }
    }
    WebGridEditDoneEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridEditDoneEventArgsDetailDescription, 'WebGridEditDoneEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebGridEditDoneEventArgsDetailDescription.__marshalByValue1 = true;
    WebGridEditDoneEventArgsDetailDescription.__marshalByValueAlias1 = "GridEditDoneEventArgsDetail";
    return WebGridEditDoneEventArgsDetailDescription;
})();
