// igniteui-react-core/esm2015/lib/WebGroupByRecordDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGroupByRecordDescription = /*@__PURE__*/ (() => {
    class WebGroupByRecordDescription extends Description {
        get_type() {
            return "WebGroupByRecord";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.q = 0;
            this.v = null;
            this.t = null;
            this.l = null;
            this.h = null;
            this.p = 0;
            this.j = null;
        }
        get expression() {
            return this.n;
        }
        set expression(a) {
            this.n = a;
            this.g("Expression");
        }
        get level() {
            return this.q;
        }
        set level(a) {
            this.q = a;
            this.g("Level");
        }
        get recordsRef() {
            return this.v;
        }
        set recordsRef(a) {
            this.v = a;
            this.g("RecordsRef");
        }
        get value() {
            return this.t;
        }
        set value(a) {
            this.t = a;
            this.g("Value");
        }
        get groupParent() {
            return this.l;
        }
        set groupParent(a) {
            this.l = a;
            this.g("GroupParent");
        }
        get groups() {
            return this.h;
        }
        set groups(a) {
            this.h = a;
            this.g("Groups");
        }
        get height() {
            return this.p;
        }
        set height(a) {
            this.p = a;
            this.g("Height");
        }
        get column() {
            return this.j;
        }
        set column(a) {
            this.j = a;
            this.g("Column");
        }
    }
    WebGroupByRecordDescription.$t = /*@__PURE__*/ markType(WebGroupByRecordDescription, 'WebGroupByRecordDescription', Description.$);
    return WebGroupByRecordDescription;
})();
