// igniteui-react-core/esm2015/lib/WebGroupingStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGroupingStateDescription = /*@__PURE__*/ (() => {
    class WebGroupingStateDescription extends Description {
        get_type() {
            return "WebGroupingState";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
            this.m = false;
        }
        get expressions() {
            return this.i;
        }
        set expressions(a) {
            this.i = a;
            this.g("Expressions");
        }
        get expansion() {
            return this.h;
        }
        set expansion(a) {
            this.h = a;
            this.g("Expansion");
        }
        get defaultExpanded() {
            return this.m;
        }
        set defaultExpanded(a) {
            this.m = a;
            this.g("DefaultExpanded");
        }
    }
    WebGroupingStateDescription.$t = /*@__PURE__*/ markType(WebGroupingStateDescription, 'WebGroupingStateDescription', Description.$);
    WebGroupingStateDescription.__marshalByValue = true;
    WebGroupingStateDescription.__marshalByValueAlias = "GroupingState";
    return WebGroupingStateDescription;
})();
