// igniteui-react-core/esm2015/lib/WebGridValidationStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridValidationStateDescription = /*@__PURE__*/ (() => {
    class WebGridValidationStateDescription extends Description {
        get_type() {
            return "WebGridValidationState";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = null;
        }
        get status() {
            return this.j;
        }
        set status(a) {
            this.j = a;
            this.g("Status");
        }
        get errors() {
            return this.h;
        }
        set errors(a) {
            this.h = a;
            this.g("Errors");
        }
    }
    WebGridValidationStateDescription.$t = /*@__PURE__*/ markType(WebGridValidationStateDescription, 'WebGridValidationStateDescription', Description.$);
    return WebGridValidationStateDescription;
})();
