// igniteui-react-core/esm2015/lib/WebGridToolbarActionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebGridToolbarContentDescription } from "./WebGridToolbarContentDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridToolbarActionsDescription = /*@__PURE__*/ (() => {
    class WebGridToolbarActionsDescription extends WebGridToolbarContentDescription {
        get_type() {
            return "WebGridToolbarActions";
        }
        constructor() {
            super();
            this.k = null;
            this.m = null;
        }
        get actions() {
            return this.k;
        }
        set actions(a) {
            this.k = a;
            this.g("Actions");
        }
        get name() {
            return this.m;
        }
        set name(a) {
            this.m = a;
            this.g("Name");
        }
    }
    WebGridToolbarActionsDescription.$t = /*@__PURE__*/ markType(WebGridToolbarActionsDescription, 'WebGridToolbarActionsDescription', WebGridToolbarContentDescription.$);
    return WebGridToolbarActionsDescription;
})();
