// igniteui-react-core/esm2015/lib/WebGridStateOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridStateOptionsDescription = /*@__PURE__*/ (() => {
    class WebGridStateOptionsDescription extends Description {
        get_type() {
            return "WebGridStateOptions";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = false;
            this.m = false;
            this.h = false;
            this.v = false;
            this.n = false;
            this.p = false;
            this.i = false;
            this.u = false;
            this.k = false;
            this.t = false;
            this.q = false;
            this.l = false;
            this.s = false;
            this.o = false;
            this.r = false;
        }
        get columns() {
            return this.j;
        }
        set columns(a) {
            this.j = a;
            this.g("Columns");
        }
        get filtering() {
            return this.m;
        }
        set filtering(a) {
            this.m = a;
            this.g("Filtering");
        }
        get advancedFiltering() {
            return this.h;
        }
        set advancedFiltering(a) {
            this.h = a;
            this.g("AdvancedFiltering");
        }
        get sorting() {
            return this.v;
        }
        set sorting(a) {
            this.v = a;
            this.g("Sorting");
        }
        get groupBy() {
            return this.n;
        }
        set groupBy(a) {
            this.n = a;
            this.g("GroupBy");
        }
        get paging() {
            return this.p;
        }
        set paging(a) {
            this.p = a;
            this.g("Paging");
        }
        get cellSelection() {
            return this.i;
        }
        set cellSelection(a) {
            this.i = a;
            this.g("CellSelection");
        }
        get rowSelection() {
            return this.u;
        }
        set rowSelection(a) {
            this.u = a;
            this.g("RowSelection");
        }
        get columnSelection() {
            return this.k;
        }
        set columnSelection(a) {
            this.k = a;
            this.g("ColumnSelection");
        }
        get rowPinning() {
            return this.t;
        }
        set rowPinning(a) {
            this.t = a;
            this.g("RowPinning");
        }
        get pinningConfig() {
            return this.q;
        }
        set pinningConfig(a) {
            this.q = a;
            this.g("PinningConfig");
        }
        get expansion() {
            return this.l;
        }
        set expansion(a) {
            this.l = a;
            this.g("Expansion");
        }
        get rowIslands() {
            return this.s;
        }
        set rowIslands(a) {
            this.s = a;
            this.g("RowIslands");
        }
        get moving() {
            return this.o;
        }
        set moving(a) {
            this.o = a;
            this.g("Moving");
        }
        get pivotConfiguration() {
            return this.r;
        }
        set pivotConfiguration(a) {
            this.r = a;
            this.g("PivotConfiguration");
        }
    }
    WebGridStateOptionsDescription.$t = /*@__PURE__*/ markType(WebGridStateOptionsDescription, 'WebGridStateOptionsDescription', Description.$);
    return WebGridStateOptionsDescription;
})();
