// igniteui-react-core/esm2015/lib/WebGroupByRowSelectorTemplateDetailsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGroupByRowSelectorTemplateDetailsDescription = /*@__PURE__*/ (() => {
    class WebGroupByRowSelectorTemplateDetailsDescription extends Description {
        get_type() {
            return "WebGroupByRowSelectorTemplateDetails";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = 0;
            this.k = 0;
            this.h = null;
        }
        get selectedCount() {
            return this.j;
        }
        set selectedCount(a) {
            this.j = a;
            this.g("SelectedCount");
        }
        get totalCount() {
            return this.k;
        }
        set totalCount(a) {
            this.k = a;
            this.g("TotalCount");
        }
        get groupRow() {
            return this.h;
        }
        set groupRow(a) {
            this.h = a;
            this.g("GroupRow");
        }
    }
    WebGroupByRowSelectorTemplateDetailsDescription.$t = /*@__PURE__*/ markType(WebGroupByRowSelectorTemplateDetailsDescription, 'WebGroupByRowSelectorTemplateDetailsDescription', Description.$);
    return WebGroupByRowSelectorTemplateDetailsDescription;
})();
