// igniteui-react-core/esm2015/lib/WebGridRowEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridRowEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridRowEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebGridRowEventArgsDetail";
        }
        constructor() {
            super();
            this.m = null;
            this.p = null;
        }
        get row() {
            return this.m;
        }
        set row(a) {
            this.m = a;
            this.g("Row");
        }
        get event() {
            return this.p;
        }
        set event(a) {
            this.p = a;
            this.g("Event");
        }
    }
    WebGridRowEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridRowEventArgsDetailDescription, 'WebGridRowEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebGridRowEventArgsDetailDescription.__marshalByValue1 = true;
    WebGridRowEventArgsDetailDescription.__marshalByValueAlias1 = "GridRowEventArgsDetail";
    return WebGridRowEventArgsDetailDescription;
})();
