// igniteui-react-core/esm2015/lib/WebGridRowDragGhostContextDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridRowDragGhostContextDescription = /*@__PURE__*/ (() => {
    class WebGridRowDragGhostContextDescription extends Description {
        get_type() {
            return "WebGridRowDragGhostContext";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.k = null;
            this.h = null;
        }
        get implicit() {
            return this.l;
        }
        set implicit(a) {
            this.l = a;
            this.g("Implicit");
        }
        get data() {
            return this.k;
        }
        set data(a) {
            this.k = a;
            this.g("Data");
        }
        get grid() {
            return this.h;
        }
        set grid(a) {
            this.h = a;
            this.g("Grid");
        }
    }
    WebGridRowDragGhostContextDescription.$t = /*@__PURE__*/ markType(WebGridRowDragGhostContextDescription, 'WebGridRowDragGhostContextDescription', Description.$);
    WebGridRowDragGhostContextDescription.__marshalByValue = true;
    WebGridRowDragGhostContextDescription.__marshalByValueAlias = "GridRowDragGhostContext";
    return WebGridRowDragGhostContextDescription;
})();
