// igniteui-react-core/esm2015/lib/WebGridEditingActionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebGridActionsBaseDirectiveDescription } from "./WebGridActionsBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridEditingActionsDescription = /*@__PURE__*/ (() => {
    class WebGridEditingActionsDescription extends WebGridActionsBaseDirectiveDescription {
        get_type() {
            return "WebGridEditingActions";
        }
        constructor() {
            super();
            this.n = false;
            this.p = false;
            this.o = false;
            this.m = false;
        }
        get addRow() {
            return this.n;
        }
        set addRow(a) {
            this.n = a;
            this.g("AddRow");
        }
        get editRow() {
            return this.p;
        }
        set editRow(a) {
            this.p = a;
            this.g("EditRow");
        }
        get deleteRow() {
            return this.o;
        }
        set deleteRow(a) {
            this.o = a;
            this.g("DeleteRow");
        }
        get addChild() {
            return this.m;
        }
        set addChild(a) {
            this.m = a;
            this.g("AddChild");
        }
    }
    WebGridEditingActionsDescription.$t = /*@__PURE__*/ markType(WebGridEditingActionsDescription, 'WebGridEditingActionsDescription', WebGridActionsBaseDirectiveDescription.$);
    return WebGridEditingActionsDescription;
})();
