// igniteui-react-core/esm2015/lib/WebGridMasterDetailContextDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridMasterDetailContextDescription = /*@__PURE__*/ (() => {
    class WebGridMasterDetailContextDescription extends Description {
        get_type() {
            return "WebGridMasterDetailContext";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = null;
            this.i = 0;
        }
        get implicit() {
            return this.k;
        }
        set implicit(a) {
            this.k = a;
            this.g("Implicit");
        }
        get index() {
            return this.i;
        }
        set index(a) {
            this.i = a;
            this.g("Index");
        }
    }
    WebGridMasterDetailContextDescription.$t = /*@__PURE__*/ markType(WebGridMasterDetailContextDescription, 'WebGridMasterDetailContextDescription', Description.$);
    WebGridMasterDetailContextDescription.__marshalByValue = true;
    WebGridMasterDetailContextDescription.__marshalByValueAlias = "GridMasterDetailContext";
    return WebGridMasterDetailContextDescription;
})();
