// igniteui-react-core/esm2015/lib/WebGridEditEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridEditEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridEditEventArgsDetailDescription extends Description {
        get_type() {
            return "WebGridEditEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.o = false;
            this.y = null;
            this.w = null;
            this.z = null;
            this.l = null;
            this.x = null;
            this.v = null;
            this.u = null;
            this.h = null;
            this.j = null;
            this.p = false;
            this.q = false;
        }
        get cancel() {
            return this.o;
        }
        set cancel(a) {
            this.o = a;
            this.g("Cancel");
        }
        get rowID() {
            return this.y;
        }
        set rowID(a) {
            this.y = a;
            this.g("RowID");
        }
        get primaryKey() {
            return this.w;
        }
        set primaryKey(a) {
            this.w = a;
            this.g("PrimaryKey");
        }
        get rowKey() {
            return this.z;
        }
        set rowKey(a) {
            this.z = a;
            this.g("RowKey");
        }
        get cellID() {
            return this.l;
        }
        set cellID(a) {
            this.l = a;
            this.g("CellID");
        }
        get rowData() {
            return this.x;
        }
        set rowData(a) {
            this.x = a;
            this.g("RowData");
        }
        get oldValue() {
            return this.v;
        }
        set oldValue(a) {
            this.v = a;
            this.g("OldValue");
        }
        get newValue() {
            return this.u;
        }
        set newValue(a) {
            this.u = a;
            this.g("NewValue");
        }
        get column() {
            return this.h;
        }
        set column(a) {
            this.h = a;
            this.g("Column");
        }
        get owner() {
            return this.j;
        }
        set owner(a) {
            this.j = a;
            this.g("Owner");
        }
        get isAddRow() {
            return this.p;
        }
        set isAddRow(a) {
            this.p = a;
            this.g("IsAddRow");
        }
        get valid() {
            return this.q;
        }
        set valid(a) {
            this.q = a;
            this.g("Valid");
        }
    }
    WebGridEditEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridEditEventArgsDetailDescription, 'WebGridEditEventArgsDetailDescription', Description.$);
    WebGridEditEventArgsDetailDescription.__marshalByValue = true;
    WebGridEditEventArgsDetailDescription.__marshalByValueAlias = "GridEditEventArgsDetail";
    return WebGridEditEventArgsDetailDescription;
})();
