// igniteui-react-core/esm2015/lib/WebGridToolbarExporterDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseToolbarDirectiveDescription } from "./WebBaseToolbarDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridToolbarExporterDescription = /*@__PURE__*/ (() => {
    class WebGridToolbarExporterDescription extends WebBaseToolbarDirectiveDescription {
        get_type() {
            return "WebGridToolbarExporter";
        }
        constructor() {
            super();
            this.ae = false;
            this.af = false;
            this.ak = null;
            this.aj = null;
            this.ai = null;
        }
        get exportCSV() {
            return this.ae;
        }
        set exportCSV(a) {
            this.ae = a;
            this.g("ExportCSV");
        }
        get exportExcel() {
            return this.af;
        }
        set exportExcel(a) {
            this.af = a;
            this.g("ExportExcel");
        }
        get filename() {
            return this.ak;
        }
        set filename(a) {
            this.ak = a;
            this.g("Filename");
        }
        get exportStartedRef() {
            return this.aj;
        }
        set exportStartedRef(a) {
            this.aj = a;
            this.g("ExportStartedRef");
        }
        get exportEndedRef() {
            return this.ai;
        }
        set exportEndedRef(a) {
            this.ai = a;
            this.g("ExportEndedRef");
        }
    }
    WebGridToolbarExporterDescription.$t = /*@__PURE__*/ markType(WebGridToolbarExporterDescription, 'WebGridToolbarExporterDescription', WebBaseToolbarDirectiveDescription.$);
    return WebGridToolbarExporterDescription;
})();
