// igniteui-react-core/esm2015/lib/WebGridValidationStatusEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridValidationStatusEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridValidationStatusEventArgsDetailDescription extends Description {
        get_type() {
            return "WebGridValidationStatusEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.h = null;
        }
        get status() {
            return this.l;
        }
        set status(a) {
            this.l = a;
            this.g("Status");
        }
        get owner() {
            return this.h;
        }
        set owner(a) {
            this.h = a;
            this.g("Owner");
        }
    }
    WebGridValidationStatusEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridValidationStatusEventArgsDetailDescription, 'WebGridValidationStatusEventArgsDetailDescription', Description.$);
    WebGridValidationStatusEventArgsDetailDescription.__marshalByValue = true;
    WebGridValidationStatusEventArgsDetailDescription.__marshalByValueAlias = "GridValidationStatusEventArgsDetail";
    return WebGridValidationStatusEventArgsDetailDescription;
})();
