// igniteui-react-core/esm2015/lib/WebGridCreatedEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridCreatedEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebGridCreatedEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebGridCreatedEventArgsDetail";
        }
        constructor() {
            super();
            this.o = null;
            this.r = null;
            this.m = null;
        }
        get owner() {
            return this.o;
        }
        set owner(a) {
            this.o = a;
            this.g("Owner");
        }
        get parentID() {
            return this.r;
        }
        set parentID(a) {
            this.r = a;
            this.g("ParentID");
        }
        get grid() {
            return this.m;
        }
        set grid(a) {
            this.m = a;
            this.g("Grid");
        }
    }
    WebGridCreatedEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebGridCreatedEventArgsDetailDescription, 'WebGridCreatedEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebGridCreatedEventArgsDetailDescription.__marshalByValue1 = true;
    WebGridCreatedEventArgsDetailDescription.__marshalByValueAlias1 = "GridCreatedEventArgsDetail";
    return WebGridCreatedEventArgsDetailDescription;
})();
