// igniteui-react-core/esm2015/lib/WebGroupByExpandStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGroupByExpandStateDescription = /*@__PURE__*/ (() => {
    class WebGroupByExpandStateDescription extends Description {
        get_type() {
            return "WebGroupByExpandState";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = false;
            this.h = null;
        }
        get expanded() {
            return this.j;
        }
        set expanded(a) {
            this.j = a;
            this.g("Expanded");
        }
        get hierarchy() {
            return this.h;
        }
        set hierarchy(a) {
            this.h = a;
            this.g("Hierarchy");
        }
    }
    WebGroupByExpandStateDescription.$t = /*@__PURE__*/ markType(WebGroupByExpandStateDescription, 'WebGroupByExpandStateDescription', Description.$);
    return WebGroupByExpandStateDescription;
})();
