// igniteui-react-core/esm2015/lib/WebGridToolbarAdvancedFilteringDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebGridToolbarBaseActionDescription } from "./WebGridToolbarBaseActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebGridToolbarAdvancedFilteringDescription = /*@__PURE__*/ (() => {
    class WebGridToolbarAdvancedFilteringDescription extends WebGridToolbarBaseActionDescription {
        get_type() {
            return "WebGridToolbarAdvancedFiltering";
        }
        constructor() {
            super();
            this.k = null;
            this.m = null;
        }
        get overlaySettings() {
            return this.k;
        }
        set overlaySettings(a) {
            this.k = a;
            this.g("OverlaySettings");
        }
        get name() {
            return this.m;
        }
        set name(a) {
            this.m = a;
            this.g("Name");
        }
    }
    WebGridToolbarAdvancedFilteringDescription.$t = /*@__PURE__*/ markType(WebGridToolbarAdvancedFilteringDescription, 'WebGridToolbarAdvancedFilteringDescription', WebGridToolbarBaseActionDescription.$);
    return WebGridToolbarAdvancedFilteringDescription;
})();
