// igniteui-react-core/esm2015/lib/WebHierarchicalGridDescriptionMetadata.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, String_$type, markType } from "./type";
import { Dictionary$2 } from "./Dictionary$2";
import { WebGridToolbarDescriptionMetadata } from "./WebGridToolbarDescriptionMetadata";
import { WebGridToolbarActionsDescriptionMetadata } from "./WebGridToolbarActionsDescriptionMetadata";
import { WebGridToolbarTitleDescriptionMetadata } from "./WebGridToolbarTitleDescriptionMetadata";
import { WebGridToolbarExporterDescriptionMetadata } from "./WebGridToolbarExporterDescriptionMetadata";
import { WebGridToolbarHidingDescriptionMetadata } from "./WebGridToolbarHidingDescriptionMetadata";
import { WebGridToolbarPinningDescriptionMetadata } from "./WebGridToolbarPinningDescriptionMetadata";
import { WebGridToolbarAdvancedFilteringDescriptionMetadata } from "./WebGridToolbarAdvancedFilteringDescriptionMetadata";
import { WebGridToolbarBaseActionDescriptionMetadata } from "./WebGridToolbarBaseActionDescriptionMetadata";
import { WebGridToolbarContentDescriptionMetadata } from "./WebGridToolbarContentDescriptionMetadata";
import { WebHierarchicalGridBaseDirectiveDescriptionMetadata } from "./WebHierarchicalGridBaseDirectiveDescriptionMetadata";
import { WebHierarchicalGridDescription } from "./WebHierarchicalGridDescription";
/**
 * @hidden
 */
export let WebHierarchicalGridDescriptionMetadata = /*@__PURE__*/ (() => {
    class WebHierarchicalGridDescriptionMetadata extends Base {
        static b(a) {
            if (WebHierarchicalGridDescriptionMetadata.a == null) {
                WebHierarchicalGridDescriptionMetadata.a = new Dictionary$2(String_$type, String_$type, 0);
                WebHierarchicalGridDescriptionMetadata.c(WebHierarchicalGridDescriptionMetadata.a);
            }
            if (a.k(WebHierarchicalGridDescriptionMetadata.a)) {
                return;
            }
            a.ac(WebHierarchicalGridDescriptionMetadata.a);
            WebHierarchicalGridDescriptionMetadata.f(a);
        }
        static c(a) {
            WebHierarchicalGridBaseDirectiveDescriptionMetadata.c(a);
            a.item("__qualifiedNameTS", "String:HierarchicalGrid");
            a.item("__tagNameWC", "String:igc-hierarchical-grid");
            a.item("__packageNameWC", "String:grids");
            a.item("__skipModuleRegisterWebComponents", "Boolean");
            a.item("__importOverrideWebComponents", "String:igniteui-webcomponents-grids/grids/combined");
            a.item("__importTypesWebComponents", "String:igniteui-webcomponents-grids/grids");
            if (a.containsKey("__skipSuffix")) {
                a.removeItem("__skipSuffix");
            }
            a.item("ChildLayoutList", "Collection:WebRowIsland:RowIslandCollection:WebRowIsland");
            a.item("ChildLayoutList@ngQueryList", "childLayoutList");
            a.item("Id", "String");
            a.item("DataRef", "(w:Data,wf:Data)DataRef::object");
            a.item("TotalItemCount", "Number:double");
            a.item("ExpandChildren", "Boolean");
            WebHierarchicalGridDescriptionMetadata.e(a);
        }
        static d(a) {
            WebHierarchicalGridDescriptionMetadata.b(a);
            a.ae("WebHierarchicalGrid", () => new WebHierarchicalGridDescription());
            a.ad("WebHierarchicalGrid", WebHierarchicalGridDescriptionMetadata.a);
        }
        static e(a) {
            a.item("__deferAttachWC", "Boolean");
            a.item("__skipModuleRegisterWebComponents", "Boolean");
            a.item("__importOverrideWebComponents", "String:igniteui-webcomponents-grids/grids/combined");
            a.item("__importTypesWebComponents", "String:igniteui-webcomponents-grids/grids");
        }
        static f(a) {
            WebGridToolbarDescriptionMetadata.d(a);
            WebGridToolbarActionsDescriptionMetadata.d(a);
            WebGridToolbarTitleDescriptionMetadata.d(a);
            WebGridToolbarExporterDescriptionMetadata.d(a);
            WebGridToolbarHidingDescriptionMetadata.d(a);
            WebGridToolbarPinningDescriptionMetadata.d(a);
            WebGridToolbarAdvancedFilteringDescriptionMetadata.d(a);
            WebGridToolbarBaseActionDescriptionMetadata.d(a);
            WebGridToolbarContentDescriptionMetadata.d(a);
        }
    }
    WebHierarchicalGridDescriptionMetadata.$t = /*@__PURE__*/ markType(WebHierarchicalGridDescriptionMetadata, 'WebHierarchicalGridDescriptionMetadata');
    WebHierarchicalGridDescriptionMetadata.a = null;
    return WebHierarchicalGridDescriptionMetadata;
})();
