// igniteui-react-core/esm2015/lib/WebHierarchicalGridBaseDirectiveDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebGridBaseDirectiveDescription } from "./WebGridBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebHierarchicalGridBaseDirectiveDescription = /*@__PURE__*/ (() => {
    class WebHierarchicalGridBaseDirectiveDescription extends WebGridBaseDirectiveDescription {
        get_type() {
            return "WebHierarchicalGridBaseDirective";
        }
        constructor() {
            super();
            this.gu = null;
            this.gr = false;
            this.gp = null;
            this.gt = null;
        }
        get hasChildrenKey() {
            return this.gu;
        }
        set hasChildrenKey(a) {
            this.gu = a;
            this.g("HasChildrenKey");
        }
        get showExpandAll() {
            return this.gr;
        }
        set showExpandAll(a) {
            this.gr = a;
            this.g("ShowExpandAll");
        }
        get rootGrid() {
            return this.gp;
        }
        set rootGrid(a) {
            this.gp = a;
            this.g("RootGrid");
        }
        get dataPreLoadRef() {
            return this.gt;
        }
        set dataPreLoadRef(a) {
            this.gt = a;
            this.g("DataPreLoadRef");
        }
    }
    WebHierarchicalGridBaseDirectiveDescription.$t = /*@__PURE__*/ markType(WebHierarchicalGridBaseDirectiveDescription, 'WebHierarchicalGridBaseDirectiveDescription', WebGridBaseDirectiveDescription.$);
    return WebHierarchicalGridBaseDirectiveDescription;
})();
