// igniteui-react-core/esm2015/lib/WebHeaderTypeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebHeaderTypeDescription = /*@__PURE__*/ (() => {
    class WebHeaderTypeDescription extends Description {
        get_type() {
            return "WebHeaderType";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.l = false;
            this.j = false;
            this.k = false;
            this.m = false;
            this.r = null;
        }
        get column() {
            return this.h;
        }
        set column(a) {
            this.h = a;
            this.g("Column");
        }
        get sorted() {
            return this.l;
        }
        set sorted(a) {
            this.l = a;
            this.g("Sorted");
        }
        get selectable() {
            return this.j;
        }
        set selectable(a) {
            this.j = a;
            this.g("Selectable");
        }
        get selected() {
            return this.k;
        }
        set selected(a) {
            this.k = a;
            this.g("Selected");
        }
        get title() {
            return this.m;
        }
        set title(a) {
            this.m = a;
            this.g("Title");
        }
        get sortDirection() {
            return this.r;
        }
        set sortDirection(a) {
            this.r = a;
            this.g("SortDirection");
        }
    }
    WebHeaderTypeDescription.$t = /*@__PURE__*/ markType(WebHeaderTypeDescription, 'WebHeaderTypeDescription', Description.$);
    return WebHeaderTypeDescription;
})();
