// igniteui-react-core/esm2015/lib/WebHierarchicalGridBaseDirectiveDescriptionMetadata.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, String_$type, markType } from "./type";
import { Dictionary$2 } from "./Dictionary$2";
import { WebGridToolbarDescriptionMetadata } from "./WebGridToolbarDescriptionMetadata";
import { WebGridToolbarActionsDescriptionMetadata } from "./WebGridToolbarActionsDescriptionMetadata";
import { WebGridToolbarTitleDescriptionMetadata } from "./WebGridToolbarTitleDescriptionMetadata";
import { WebGridToolbarExporterDescriptionMetadata } from "./WebGridToolbarExporterDescriptionMetadata";
import { WebGridToolbarHidingDescriptionMetadata } from "./WebGridToolbarHidingDescriptionMetadata";
import { WebGridToolbarPinningDescriptionMetadata } from "./WebGridToolbarPinningDescriptionMetadata";
import { WebGridToolbarAdvancedFilteringDescriptionMetadata } from "./WebGridToolbarAdvancedFilteringDescriptionMetadata";
import { WebGridToolbarBaseActionDescriptionMetadata } from "./WebGridToolbarBaseActionDescriptionMetadata";
import { WebGridToolbarContentDescriptionMetadata } from "./WebGridToolbarContentDescriptionMetadata";
import { WebGridBaseDirectiveDescriptionMetadata } from "./WebGridBaseDirectiveDescriptionMetadata";
import { WebForOfStateEventArgsDescriptionMetadata } from "./WebForOfStateEventArgsDescriptionMetadata";
/**
 * @hidden
 */
export let WebHierarchicalGridBaseDirectiveDescriptionMetadata = /*@__PURE__*/ (() => {
    class WebHierarchicalGridBaseDirectiveDescriptionMetadata extends Base {
        static b(a) {
            if (WebHierarchicalGridBaseDirectiveDescriptionMetadata.a == null) {
                WebHierarchicalGridBaseDirectiveDescriptionMetadata.a = new Dictionary$2(String_$type, String_$type, 0);
                WebHierarchicalGridBaseDirectiveDescriptionMetadata.c(WebHierarchicalGridBaseDirectiveDescriptionMetadata.a);
            }
            if (a.k(WebHierarchicalGridBaseDirectiveDescriptionMetadata.a)) {
                return;
            }
            a.ac(WebHierarchicalGridBaseDirectiveDescriptionMetadata.a);
            WebHierarchicalGridBaseDirectiveDescriptionMetadata.f(a);
            WebGridBaseDirectiveDescriptionMetadata.d(a);
            WebForOfStateEventArgsDescriptionMetadata.d(a);
        }
        static c(a) {
            WebGridBaseDirectiveDescriptionMetadata.c(a);
            a.item("__qualifiedNameTS", "String:HierarchicalGridBaseDirective");
            a.item("__packageNameWC", "String:grids");
            a.item("__skipModuleRegisterWebComponents", "Boolean");
            a.item("__importOverrideWebComponents", "String:igniteui-webcomponents-grids/grids/combined");
            a.item("__importTypesWebComponents", "String:igniteui-webcomponents-grids/grids");
            a.item("__skipSuffix", "Boolean");
            a.item("HasChildrenKey", "String");
            a.item("ShowExpandAll", "Boolean");
            a.item("RootGrid", "ExportedType:WebGridBaseDirective");
            a.item("DataPreLoadRef", "EventRef::dataPreLoad:customEvent");
            WebHierarchicalGridBaseDirectiveDescriptionMetadata.e(a);
        }
        static d(a) {
            WebHierarchicalGridBaseDirectiveDescriptionMetadata.b(a);
            a.ad("WebHierarchicalGridBaseDirective", WebHierarchicalGridBaseDirectiveDescriptionMetadata.a);
        }
        static e(a) {
            a.item("__deferAttachWC", "Boolean");
            a.item("__skipModuleRegisterWebComponents", "Boolean");
            a.item("__importOverrideWebComponents", "String:igniteui-webcomponents-grids/grids/combined");
            a.item("__importTypesWebComponents", "String:igniteui-webcomponents-grids/grids");
        }
        static f(a) {
            WebGridToolbarDescriptionMetadata.d(a);
            WebGridToolbarActionsDescriptionMetadata.d(a);
            WebGridToolbarTitleDescriptionMetadata.d(a);
            WebGridToolbarExporterDescriptionMetadata.d(a);
            WebGridToolbarHidingDescriptionMetadata.d(a);
            WebGridToolbarPinningDescriptionMetadata.d(a);
            WebGridToolbarAdvancedFilteringDescriptionMetadata.d(a);
            WebGridToolbarBaseActionDescriptionMetadata.d(a);
            WebGridToolbarContentDescriptionMetadata.d(a);
        }
    }
    WebHierarchicalGridBaseDirectiveDescriptionMetadata.$t = /*@__PURE__*/ markType(WebHierarchicalGridBaseDirectiveDescriptionMetadata, 'WebHierarchicalGridBaseDirectiveDescriptionMetadata');
    WebHierarchicalGridBaseDirectiveDescriptionMetadata.a = null;
    return WebHierarchicalGridBaseDirectiveDescriptionMetadata;
})();
