// igniteui-react-core/esm2015/lib/WebHierarchicalGridDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebHierarchicalGridBaseDirectiveDescription } from "./WebHierarchicalGridBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebHierarchicalGridDescription = /*@__PURE__*/ (() => {
    class WebHierarchicalGridDescription extends WebHierarchicalGridBaseDirectiveDescription {
        get_type() {
            return "WebHierarchicalGrid";
        }
        constructor() {
            super();
            this.gx = null;
            this.g4 = null;
            this.g3 = null;
            this.g1 = 0;
            this.gz = false;
        }
        get childLayoutList() {
            return this.gx;
        }
        set childLayoutList(a) {
            this.gx = a;
            this.g("ChildLayoutList");
        }
        get id() {
            return this.g4;
        }
        set id(a) {
            this.g4 = a;
            this.g("Id");
        }
        get dataRef() {
            return this.g3;
        }
        set dataRef(a) {
            this.g3 = a;
            this.g("DataRef");
        }
        get totalItemCount() {
            return this.g1;
        }
        set totalItemCount(a) {
            this.g1 = a;
            this.g("TotalItemCount");
        }
        get expandChildren() {
            return this.gz;
        }
        set expandChildren(a) {
            this.gz = a;
            this.g("ExpandChildren");
        }
    }
    WebHierarchicalGridDescription.$t = /*@__PURE__*/ markType(WebHierarchicalGridDescription, 'WebHierarchicalGridDescription', WebHierarchicalGridBaseDirectiveDescription.$);
    return WebHierarchicalGridDescription;
})();
