// igniteui-react-core/esm2015/lib/WebInputDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebInputBaseDescription } from "./WebInputBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebInputDescription = /*@__PURE__*/ (() => {
    class WebInputDescription extends WebInputBaseDescription {
        get_type() {
            return "WebInput";
        }
        constructor() {
            super();
            this.ax = null;
            this.as = null;
            this.at = null;
            this.aw = null;
            this.al = 0;
            this.ak = 0;
            this.av = null;
            this.au = null;
            this.am = 0;
            this.ag = false;
            this.aq = null;
            this.ah = false;
            this.ar = null;
        }
        get value() {
            return this.ax;
        }
        set value(a) {
            this.ax = a;
            this.g("Value");
        }
        get displayType() {
            return this.as;
        }
        set displayType(a) {
            this.as = a;
            this.g("DisplayType");
        }
        get inputMode() {
            return this.at;
        }
        set inputMode(a) {
            this.at = a;
            this.g("InputMode");
        }
        get pattern() {
            return this.aw;
        }
        set pattern(a) {
            this.aw = a;
            this.g("Pattern");
        }
        get minLength() {
            return this.al;
        }
        set minLength(a) {
            this.al = a;
            this.g("MinLength");
        }
        get maxLength() {
            return this.ak;
        }
        set maxLength(a) {
            this.ak = a;
            this.g("MaxLength");
        }
        get min() {
            return this.av;
        }
        set min(a) {
            this.av = a;
            this.g("Min");
        }
        get max() {
            return this.au;
        }
        set max(a) {
            this.au = a;
            this.g("Max");
        }
        get step() {
            return this.am;
        }
        set step(a) {
            this.am = a;
            this.g("Step");
        }
        get autofocus() {
            return this.ag;
        }
        set autofocus(a) {
            this.ag = a;
            this.g("Autofocus");
        }
        get autocomplete() {
            return this.aq;
        }
        set autocomplete(a) {
            this.aq = a;
            this.g("Autocomplete");
        }
        get validateOnly() {
            return this.ah;
        }
        set validateOnly(a) {
            this.ah = a;
            this.g("ValidateOnly");
        }
        get changeRef() {
            return this.ar;
        }
        set changeRef(a) {
            this.ar = a;
            this.g("ChangeRef");
        }
    }
    WebInputDescription.$t = /*@__PURE__*/ markType(WebInputDescription, 'WebInputDescription', WebInputBaseDescription.$);
    return WebInputDescription;
})();
