// igniteui-react-core/esm2015/lib/WebLinearProgressDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebProgressBaseDescription } from "./WebProgressBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebLinearProgressDescription = /*@__PURE__*/ (() => {
    class WebLinearProgressDescription extends WebProgressBaseDescription {
        get_type() {
            return "WebLinearProgress";
        }
        constructor() {
            super();
            this.w = false;
            this.y = null;
        }
        get striped() {
            return this.w;
        }
        set striped(a) {
            this.w = a;
            this.g("Striped");
        }
        get labelAlign() {
            return this.y;
        }
        set labelAlign(a) {
            this.y = a;
            this.g("LabelAlign");
        }
    }
    WebLinearProgressDescription.$t = /*@__PURE__*/ markType(WebLinearProgressDescription, 'WebLinearProgressDescription', WebProgressBaseDescription.$);
    return WebLinearProgressDescription;
})();
