// igniteui-react-core/esm2015/lib/WebMoveTabActionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebPaneDragActionDescription } from "./WebPaneDragActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebMoveTabActionDescription = /*@__PURE__*/ (() => {
    class WebMoveTabActionDescription extends WebPaneDragActionDescription {
        get_type() {
            return "WebMoveTabAction";
        }
        constructor() {
            super();
            this.q = null;
            this.m = 0;
            this.l = 0;
        }
        get actionType() {
            return this.q;
        }
        set actionType(a) {
            this.q = a;
            this.g("ActionType");
        }
        get oldIndex() {
            return this.m;
        }
        set oldIndex(a) {
            this.m = a;
            this.g("OldIndex");
        }
        get newIndex() {
            return this.l;
        }
        set newIndex(a) {
            this.l = a;
            this.g("NewIndex");
        }
    }
    WebMoveTabActionDescription.$t = /*@__PURE__*/ markType(WebMoveTabActionDescription, 'WebMoveTabActionDescription', WebPaneDragActionDescription.$);
    WebMoveTabActionDescription.__marshalByValue1 = true;
    WebMoveTabActionDescription.__marshalByValueAlias1 = "MoveTabAction";
    return WebMoveTabActionDescription;
})();
