// igniteui-react-core/esm2015/lib/WebMoveFloatingPaneActionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebPaneDragActionDescription } from "./WebPaneDragActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebMoveFloatingPaneActionDescription = /*@__PURE__*/ (() => {
    class WebMoveFloatingPaneActionDescription extends WebPaneDragActionDescription {
        get_type() {
            return "WebMoveFloatingPaneAction";
        }
        constructor() {
            super();
            this.q = null;
            this.l = null;
            this.k = null;
        }
        get actionType() {
            return this.q;
        }
        set actionType(a) {
            this.q = a;
            this.g("ActionType");
        }
        get oldLocation() {
            return this.l;
        }
        set oldLocation(a) {
            this.l = a;
            this.g("OldLocation");
        }
        get newLocation() {
            return this.k;
        }
        set newLocation(a) {
            this.k = a;
            this.g("NewLocation");
        }
    }
    WebMoveFloatingPaneActionDescription.$t = /*@__PURE__*/ markType(WebMoveFloatingPaneActionDescription, 'WebMoveFloatingPaneActionDescription', WebPaneDragActionDescription.$);
    WebMoveFloatingPaneActionDescription.__marshalByValue1 = true;
    WebMoveFloatingPaneActionDescription.__marshalByValueAlias1 = "MoveFloatingPaneAction";
    return WebMoveFloatingPaneActionDescription;
})();
