// igniteui-react-core/esm2015/lib/WebMaskInputDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebMaskInputBaseDescription } from "./WebMaskInputBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebMaskInputDescription = /*@__PURE__*/ (() => {
    class WebMaskInputDescription extends WebMaskInputBaseDescription {
        get_type() {
            return "WebMaskInput";
        }
        constructor() {
            super();
            this.al = null;
            this.ak = null;
            this.aj = null;
            this.ai = null;
        }
        get valueMode() {
            return this.al;
        }
        set valueMode(a) {
            this.al = a;
            this.g("ValueMode");
        }
        get value() {
            return this.ak;
        }
        set value(a) {
            this.ak = a;
            this.g("Value");
        }
        get mask() {
            return this.aj;
        }
        set mask(a) {
            this.aj = a;
            this.g("Mask");
        }
        get changeRef() {
            return this.ai;
        }
        set changeRef(a) {
            this.ai = a;
            this.g("ChangeRef");
        }
    }
    WebMaskInputDescription.$t = /*@__PURE__*/ markType(WebMaskInputDescription, 'WebMaskInputDescription', WebMaskInputBaseDescription.$);
    return WebMaskInputDescription;
})();
