// igniteui-react-core/esm2015/lib/WebNavDrawerItemDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebNavDrawerItemDescription = /*@__PURE__*/ (() => {
    class WebNavDrawerItemDescription extends Description {
        get_type() {
            return "WebNavDrawerItem";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = false;
            this.h = false;
        }
        get disabled() {
            return this.i;
        }
        set disabled(a) {
            this.i = a;
            this.g("Disabled");
        }
        get active() {
            return this.h;
        }
        set active(a) {
            this.h = a;
            this.g("Active");
        }
    }
    WebNavDrawerItemDescription.$t = /*@__PURE__*/ markType(WebNavDrawerItemDescription, 'WebNavDrawerItemDescription', Description.$);
    return WebNavDrawerItemDescription;
})();
