// igniteui-react-core/esm2015/lib/WebNavDrawerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebNavDrawerDescription = /*@__PURE__*/ (() => {
    class WebNavDrawerDescription extends Description {
        get_type() {
            return "WebNavDrawer";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = false;
        }
        get position() {
            return this.j;
        }
        set position(a) {
            this.j = a;
            this.g("Position");
        }
        get open() {
            return this.h;
        }
        set open(a) {
            this.h = a;
            this.g("Open");
        }
    }
    WebNavDrawerDescription.$t = /*@__PURE__*/ markType(WebNavDrawerDescription, 'WebNavDrawerDescription', Description.$);
    return WebNavDrawerDescription;
})();
