// igniteui-react-core/esm2015/lib/WebOverlaySettingsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebOverlaySettingsDescription = /*@__PURE__*/ (() => {
    class WebOverlaySettingsDescription extends Description {
        get_type() {
            return "WebOverlaySettings";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.r = null;
            this.h = null;
            this.j = null;
            this.n = false;
            this.m = false;
            this.l = false;
        }
        get target() {
            return this.r;
        }
        set target(a) {
            this.r = a;
            this.g("Target");
        }
        get positionStrategy() {
            return this.h;
        }
        set positionStrategy(a) {
            this.h = a;
            this.g("PositionStrategy");
        }
        get scrollStrategy() {
            return this.j;
        }
        set scrollStrategy(a) {
            this.j = a;
            this.g("ScrollStrategy");
        }
        get modal() {
            return this.n;
        }
        set modal(a) {
            this.n = a;
            this.g("Modal");
        }
        get closeOnOutsideClick() {
            return this.m;
        }
        set closeOnOutsideClick(a) {
            this.m = a;
            this.g("CloseOnOutsideClick");
        }
        get closeOnEscape() {
            return this.l;
        }
        set closeOnEscape(a) {
            this.l = a;
            this.g("CloseOnEscape");
        }
    }
    WebOverlaySettingsDescription.$t = /*@__PURE__*/ markType(WebOverlaySettingsDescription, 'WebOverlaySettingsDescription', Description.$);
    return WebOverlaySettingsDescription;
})();
