// igniteui-react-core/esm2015/lib/WebPinRowEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPinRowEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebPinRowEventArgsDetailDescription extends Description {
        get_type() {
            return "WebPinRowEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.r = null;
            this.s = null;
            this.h = null;
            this.o = 0;
            this.l = false;
            this.q = null;
            this.k = false;
        }
        get rowID() {
            return this.r;
        }
        set rowID(a) {
            this.r = a;
            this.g("RowID");
        }
        get rowKey() {
            return this.s;
        }
        set rowKey(a) {
            this.s = a;
            this.g("RowKey");
        }
        get row() {
            return this.h;
        }
        set row(a) {
            this.h = a;
            this.g("Row");
        }
        get insertAtIndex() {
            return this.o;
        }
        set insertAtIndex(a) {
            this.o = a;
            this.g("InsertAtIndex");
        }
        get isPinned() {
            return this.l;
        }
        set isPinned(a) {
            this.l = a;
            this.g("IsPinned");
        }
        get owner() {
            return this.q;
        }
        set owner(a) {
            this.q = a;
            this.g("Owner");
        }
        get cancel() {
            return this.k;
        }
        set cancel(a) {
            this.k = a;
            this.g("Cancel");
        }
    }
    WebPinRowEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebPinRowEventArgsDetailDescription, 'WebPinRowEventArgsDetailDescription', Description.$);
    WebPinRowEventArgsDetailDescription.__marshalByValue = true;
    WebPinRowEventArgsDetailDescription.__marshalByValueAlias = "PinRowEventArgsDetail";
    return WebPinRowEventArgsDetailDescription;
})();
