// igniteui-react-core/esm2015/lib/WebPivotConfigurationDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPivotConfigurationDescription = /*@__PURE__*/ (() => {
    class WebPivotConfigurationDescription extends Description {
        get_type() {
            return "WebPivotConfiguration";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.q = null;
            this.p = null;
            this.j = null;
            this.h = null;
            this.k = null;
            this.i = null;
            this.t = null;
        }
        get rowStrategy() {
            return this.q;
        }
        set rowStrategy(a) {
            this.q = a;
            this.g("RowStrategy");
        }
        get columnStrategy() {
            return this.p;
        }
        set columnStrategy(a) {
            this.p = a;
            this.g("ColumnStrategy");
        }
        get rows() {
            return this.j;
        }
        set rows(a) {
            this.j = a;
            this.g("Rows");
        }
        get columns() {
            return this.h;
        }
        set columns(a) {
            this.h = a;
            this.g("Columns");
        }
        get values() {
            return this.k;
        }
        set values(a) {
            this.k = a;
            this.g("Values");
        }
        get filters() {
            return this.i;
        }
        set filters(a) {
            this.i = a;
            this.g("Filters");
        }
        get pivotKeys() {
            return this.t;
        }
        set pivotKeys(a) {
            this.t = a;
            this.g("PivotKeys");
        }
    }
    WebPivotConfigurationDescription.$t = /*@__PURE__*/ markType(WebPivotConfigurationDescription, 'WebPivotConfigurationDescription', Description.$);
    WebPivotConfigurationDescription.__marshalByValue = true;
    WebPivotConfigurationDescription.__marshalByValueAlias = "PivotConfiguration";
    return WebPivotConfigurationDescription;
})();
