// igniteui-react-core/esm2015/lib/WebPivotGridRecordDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPivotGridRecordDescription = /*@__PURE__*/ (() => {
    class WebPivotGridRecordDescription extends Description {
        get_type() {
            return "WebPivotGridRecord";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.k = 0;
            this.h = null;
            this.o = null;
            this.j = 0;
        }
        get recordsRef() {
            return this.n;
        }
        set recordsRef(a) {
            this.n = a;
            this.g("RecordsRef");
        }
        get level() {
            return this.k;
        }
        set level(a) {
            this.k = a;
            this.g("Level");
        }
        get dimensions() {
            return this.h;
        }
        set dimensions(a) {
            this.h = a;
            this.g("Dimensions");
        }
        get totalRecordDimensionName() {
            return this.o;
        }
        set totalRecordDimensionName(a) {
            this.o = a;
            this.g("TotalRecordDimensionName");
        }
        get dataIndex() {
            return this.j;
        }
        set dataIndex(a) {
            this.j = a;
            this.g("DataIndex");
        }
    }
    WebPivotGridRecordDescription.$t = /*@__PURE__*/ markType(WebPivotGridRecordDescription, 'WebPivotGridRecordDescription', Description.$);
    return WebPivotGridRecordDescription;
})();
