// igniteui-react-core/esm2015/lib/WebPaneHeaderConnectionEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPaneHeaderConnectionEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebPaneHeaderConnectionEventArgsDetailDescription extends Description {
        get_type() {
            return "WebPaneHeaderConnectionEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.j = null;
        }
        get pane() {
            return this.h;
        }
        set pane(a) {
            this.h = a;
            this.g("Pane");
        }
        get element() {
            return this.j;
        }
        set element(a) {
            this.j = a;
            this.g("Element");
        }
    }
    WebPaneHeaderConnectionEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebPaneHeaderConnectionEventArgsDetailDescription, 'WebPaneHeaderConnectionEventArgsDetailDescription', Description.$);
    WebPaneHeaderConnectionEventArgsDetailDescription.__marshalByValue = true;
    WebPaneHeaderConnectionEventArgsDetailDescription.__marshalByValueAlias = "PaneHeaderConnectionEventArgsDetail";
    return WebPaneHeaderConnectionEventArgsDetailDescription;
})();
