// igniteui-react-core/esm2015/lib/WebPaginatorResourceStringsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPaginatorResourceStringsDescription = /*@__PURE__*/ (() => {
    class WebPaginatorResourceStringsDescription extends Description {
        get_type() {
            return "WebPaginatorResourceStrings";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.l = null;
            this.h = null;
            this.m = null;
            this.j = null;
            this.k = null;
        }
        get igx_paginator_label() {
            return this.i;
        }
        set igx_paginator_label(a) {
            this.i = a;
            this.g("Igx_paginator_label");
        }
        get igx_paginator_pager_text() {
            return this.l;
        }
        set igx_paginator_pager_text(a) {
            this.l = a;
            this.g("Igx_paginator_pager_text");
        }
        get igx_paginator_first_page_button_text() {
            return this.h;
        }
        set igx_paginator_first_page_button_text(a) {
            this.h = a;
            this.g("Igx_paginator_first_page_button_text");
        }
        get igx_paginator_previous_page_button_text() {
            return this.m;
        }
        set igx_paginator_previous_page_button_text(a) {
            this.m = a;
            this.g("Igx_paginator_previous_page_button_text");
        }
        get igx_paginator_last_page_button_text() {
            return this.j;
        }
        set igx_paginator_last_page_button_text(a) {
            this.j = a;
            this.g("Igx_paginator_last_page_button_text");
        }
        get igx_paginator_next_page_button_text() {
            return this.k;
        }
        set igx_paginator_next_page_button_text(a) {
            this.k = a;
            this.g("Igx_paginator_next_page_button_text");
        }
    }
    WebPaginatorResourceStringsDescription.$t = /*@__PURE__*/ markType(WebPaginatorResourceStringsDescription, 'WebPaginatorResourceStringsDescription', Description.$);
    return WebPaginatorResourceStringsDescription;
})();
