// igniteui-react-core/esm2015/lib/WebPaneDragStartEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPaneDragStartEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebPaneDragStartEventArgsDetailDescription extends Description {
        get_type() {
            return "WebPaneDragStartEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = null;
        }
        get sourcePane() {
            return this.j;
        }
        set sourcePane(a) {
            this.j = a;
            this.g("SourcePane");
        }
        get panes() {
            return this.h;
        }
        set panes(a) {
            this.h = a;
            this.g("Panes");
        }
    }
    WebPaneDragStartEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebPaneDragStartEventArgsDetailDescription, 'WebPaneDragStartEventArgsDetailDescription', Description.$);
    WebPaneDragStartEventArgsDetailDescription.__marshalByValue = true;
    WebPaneDragStartEventArgsDetailDescription.__marshalByValueAlias = "PaneDragStartEventArgsDetail";
    return WebPaneDragStartEventArgsDetailDescription;
})();
