// igniteui-react-core/esm2015/lib/WebPivotDataSelectorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPivotDataSelectorDescription = /*@__PURE__*/ (() => {
    class WebPivotDataSelectorDescription extends Description {
        get_type() {
            return "WebPivotDataSelector";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = false;
            this.j = false;
            this.i = false;
            this.k = false;
            this.p = null;
        }
        get columnsExpanded() {
            return this.h;
        }
        set columnsExpanded(a) {
            this.h = a;
            this.g("ColumnsExpanded");
        }
        get rowsExpanded() {
            return this.j;
        }
        set rowsExpanded(a) {
            this.j = a;
            this.g("RowsExpanded");
        }
        get filtersExpanded() {
            return this.i;
        }
        set filtersExpanded(a) {
            this.i = a;
            this.g("FiltersExpanded");
        }
        get valuesExpanded() {
            return this.k;
        }
        set valuesExpanded(a) {
            this.k = a;
            this.g("ValuesExpanded");
        }
        get gridRef() {
            return this.p;
        }
        set gridRef(a) {
            this.p = a;
            this.g("GridRef");
        }
    }
    WebPivotDataSelectorDescription.$t = /*@__PURE__*/ markType(WebPivotDataSelectorDescription, 'WebPivotDataSelectorDescription', Description.$);
    return WebPivotDataSelectorDescription;
})();
