// igniteui-react-core/esm2015/lib/WebPinColumnCancellableEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPinColumnCancellableEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebPinColumnCancellableEventArgsDetailDescription extends Description {
        get_type() {
            return "WebPinColumnCancellableEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.o = 0;
            this.l = false;
            this.k = false;
        }
        get column() {
            return this.h;
        }
        set column(a) {
            this.h = a;
            this.g("Column");
        }
        get insertAtIndex() {
            return this.o;
        }
        set insertAtIndex(a) {
            this.o = a;
            this.g("InsertAtIndex");
        }
        get isPinned() {
            return this.l;
        }
        set isPinned(a) {
            this.l = a;
            this.g("IsPinned");
        }
        get cancel() {
            return this.k;
        }
        set cancel(a) {
            this.k = a;
            this.g("Cancel");
        }
    }
    WebPinColumnCancellableEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebPinColumnCancellableEventArgsDetailDescription, 'WebPinColumnCancellableEventArgsDetailDescription', Description.$);
    WebPinColumnCancellableEventArgsDetailDescription.__marshalByValue = true;
    WebPinColumnCancellableEventArgsDetailDescription.__marshalByValueAlias = "PinColumnCancellableEventArgsDetail";
    return WebPinColumnCancellableEventArgsDetailDescription;
})();
