// igniteui-react-core/esm2015/lib/WebPivotValueDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPivotValueDetailDescription = /*@__PURE__*/ (() => {
    class WebPivotValueDetailDescription extends Description {
        get_type() {
            return "WebPivotValueDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.u = null;
            this.s = null;
            this.j = null;
            this.h = null;
            this.m = false;
            this.o = null;
            this.r = null;
            this.t = null;
        }
        get member() {
            return this.u;
        }
        set member(a) {
            this.u = a;
            this.g("Member");
        }
        get displayName() {
            return this.s;
        }
        set displayName(a) {
            this.s = a;
            this.g("DisplayName");
        }
        get aggregate() {
            return this.j;
        }
        set aggregate(a) {
            this.j = a;
            this.g("Aggregate");
        }
        get aggregateList() {
            return this.h;
        }
        set aggregateList(a) {
            this.h = a;
            this.g("AggregateList");
        }
        get enabled() {
            return this.m;
        }
        set enabled(a) {
            this.m = a;
            this.g("Enabled");
        }
        get styles() {
            return this.o;
        }
        set styles(a) {
            this.o = a;
            this.g("Styles");
        }
        get dataType() {
            return this.r;
        }
        set dataType(a) {
            this.r = a;
            this.g("DataType");
        }
        get formatterRef() {
            return this.t;
        }
        set formatterRef(a) {
            this.t = a;
            this.g("FormatterRef");
        }
    }
    WebPivotValueDetailDescription.$t = /*@__PURE__*/ markType(WebPivotValueDetailDescription, 'WebPivotValueDetailDescription', Description.$);
    WebPivotValueDetailDescription.__marshalByValue = true;
    WebPivotValueDetailDescription.__marshalByValueAlias = "PivotValueDetail";
    return WebPivotValueDetailDescription;
})();
