// igniteui-react-core/esm2015/lib/WebPaneDragOverEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPaneDragOverEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebPaneDragOverEventArgsDetailDescription extends Description {
        get_type() {
            return "WebPaneDragOverEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = null;
            this.l = null;
            this.o = false;
        }
        get sourcePane() {
            return this.j;
        }
        set sourcePane(a) {
            this.j = a;
            this.g("SourcePane");
        }
        get panes() {
            return this.h;
        }
        set panes(a) {
            this.h = a;
            this.g("Panes");
        }
        get action() {
            return this.l;
        }
        set action(a) {
            this.l = a;
            this.g("Action");
        }
        get isValid() {
            return this.o;
        }
        set isValid(a) {
            this.o = a;
            this.g("IsValid");
        }
    }
    WebPaneDragOverEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebPaneDragOverEventArgsDetailDescription, 'WebPaneDragOverEventArgsDetailDescription', Description.$);
    WebPaneDragOverEventArgsDetailDescription.__marshalByValue = true;
    WebPaneDragOverEventArgsDetailDescription.__marshalByValueAlias = "PaneDragOverEventArgsDetail";
    return WebPaneDragOverEventArgsDetailDescription;
})();
