// igniteui-react-core/esm2015/lib/WebPagingStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPagingStateDescription = /*@__PURE__*/ (() => {
    class WebPagingStateDescription extends Description {
        get_type() {
            return "WebPagingState";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = 0;
            this.j = 0;
        }
        get index() {
            return this.i;
        }
        set index(a) {
            this.i = a;
            this.g("Index");
        }
        get recordsPerPage() {
            return this.j;
        }
        set recordsPerPage(a) {
            this.j = a;
            this.g("RecordsPerPage");
        }
    }
    WebPagingStateDescription.$t = /*@__PURE__*/ markType(WebPagingStateDescription, 'WebPagingStateDescription', Description.$);
    WebPagingStateDescription.__marshalByValue = true;
    WebPagingStateDescription.__marshalByValueAlias = "PagingState";
    return WebPagingStateDescription;
})();
