// igniteui-react-core/esm2015/lib/WebPivotDateDimensionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebPivotDimensionDescription } from "./WebPivotDimensionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPivotDateDimensionDescription = /*@__PURE__*/ (() => {
    class WebPivotDateDimensionDescription extends WebPivotDimensionDescription {
        get_type() {
            return "WebPivotDateDimension";
        }
        constructor() {
            super();
            this.ap = false;
            this.as = null;
            this.ai = null;
            this.am = null;
            this.ak = null;
            this.at = null;
        }
        get enabled() {
            return this.ap;
        }
        set enabled(a) {
            this.ap = a;
            this.g("Enabled");
        }
        get dataType() {
            return this.as;
        }
        set dataType(a) {
            this.as = a;
            this.g("DataType");
        }
        get resourceStrings() {
            return this.ai;
        }
        set resourceStrings(a) {
            this.ai = a;
            this.g("ResourceStrings");
        }
        get baseDimension() {
            return this.am;
        }
        set baseDimension(a) {
            this.am = a;
            this.g("BaseDimension");
        }
        get options() {
            return this.ak;
        }
        set options(a) {
            this.ak = a;
            this.g("Options");
        }
        get displayName() {
            return this.at;
        }
        set displayName(a) {
            this.at = a;
            this.g("DisplayName");
        }
    }
    WebPivotDateDimensionDescription.$t = /*@__PURE__*/ markType(WebPivotDateDimensionDescription, 'WebPivotDateDimensionDescription', WebPivotDimensionDescription.$);
    WebPivotDateDimensionDescription.__marshalByValue1 = true;
    WebPivotDateDimensionDescription.__marshalByValueAlias1 = "PivotDateDimension";
    return WebPivotDateDimensionDescription;
})();
