// igniteui-react-core/esm2015/lib/WebPivotDateDimensionOptionsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPivotDateDimensionOptionsDescription = /*@__PURE__*/ (() => {
    class WebPivotDateDimensionOptionsDescription extends Description {
        get_type() {
            return "WebPivotDateDimensionOptions";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = false;
            this.l = false;
            this.j = false;
            this.i = false;
            this.h = false;
        }
        get total() {
            return this.k;
        }
        set total(a) {
            this.k = a;
            this.g("Total");
        }
        get years() {
            return this.l;
        }
        set years(a) {
            this.l = a;
            this.g("Years");
        }
        get quarters() {
            return this.j;
        }
        set quarters(a) {
            this.j = a;
            this.g("Quarters");
        }
        get months() {
            return this.i;
        }
        set months(a) {
            this.i = a;
            this.g("Months");
        }
        get fullDate() {
            return this.h;
        }
        set fullDate(a) {
            this.h = a;
            this.g("FullDate");
        }
    }
    WebPivotDateDimensionOptionsDescription.$t = /*@__PURE__*/ markType(WebPivotDateDimensionOptionsDescription, 'WebPivotDateDimensionOptionsDescription', Description.$);
    return WebPivotDateDimensionOptionsDescription;
})();
