// igniteui-react-core/esm2015/lib/WebPivotDimensionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPivotDimensionDescription = /*@__PURE__*/ (() => {
    class WebPivotDimensionDescription extends Description {
        get_type() {
            return "WebPivotDimension";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.y = null;
            this.w = null;
            this.m = false;
            this.h = null;
            this.o = false;
            this.z = null;
            this.v = null;
            this.aa = null;
            this.s = 0;
            this.n = false;
            this.x = null;
        }
        get childLevel() {
            return this.j;
        }
        set childLevel(a) {
            this.j = a;
            this.g("ChildLevel");
        }
        get memberName() {
            return this.y;
        }
        set memberName(a) {
            this.y = a;
            this.g("MemberName");
        }
        get displayName() {
            return this.w;
        }
        set displayName(a) {
            this.w = a;
            this.g("DisplayName");
        }
        get enabled() {
            return this.m;
        }
        set enabled(a) {
            this.m = a;
            this.g("Enabled");
        }
        get filter() {
            return this.h;
        }
        set filter(a) {
            this.h = a;
            this.g("Filter");
        }
        get sortable() {
            return this.o;
        }
        set sortable(a) {
            this.o = a;
            this.g("Sortable");
        }
        get sortDirection() {
            return this.z;
        }
        set sortDirection(a) {
            this.z = a;
            this.g("SortDirection");
        }
        get dataType() {
            return this.v;
        }
        set dataType(a) {
            this.v = a;
            this.g("DataType");
        }
        get width() {
            return this.aa;
        }
        set width(a) {
            this.aa = a;
            this.g("Width");
        }
        get level() {
            return this.s;
        }
        set level(a) {
            this.s = a;
            this.g("Level");
        }
        get horizontalSummary() {
            return this.n;
        }
        set horizontalSummary(a) {
            this.n = a;
            this.g("HorizontalSummary");
        }
        get memberFunctionRef() {
            return this.x;
        }
        set memberFunctionRef(a) {
            this.x = a;
            this.g("MemberFunctionRef");
        }
    }
    WebPivotDimensionDescription.$t = /*@__PURE__*/ markType(WebPivotDimensionDescription, 'WebPivotDimensionDescription', Description.$);
    WebPivotDimensionDescription.__marshalByValue = true;
    WebPivotDimensionDescription.__marshalByValueAlias = "PivotDimension";
    return WebPivotDimensionDescription;
})();
