// igniteui-react-core/esm2015/lib/WebPaneDragEndEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPaneDragEndEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebPaneDragEndEventArgsDetailDescription extends Description {
        get_type() {
            return "WebPaneDragEndEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.h = null;
        }
        get sourcePane() {
            return this.j;
        }
        set sourcePane(a) {
            this.j = a;
            this.g("SourcePane");
        }
        get panes() {
            return this.h;
        }
        set panes(a) {
            this.h = a;
            this.g("Panes");
        }
    }
    WebPaneDragEndEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebPaneDragEndEventArgsDetailDescription, 'WebPaneDragEndEventArgsDetailDescription', Description.$);
    WebPaneDragEndEventArgsDetailDescription.__marshalByValue = true;
    WebPaneDragEndEventArgsDetailDescription.__marshalByValueAlias = "PaneDragEndEventArgsDetail";
    return WebPaneDragEndEventArgsDetailDescription;
})();
