// igniteui-react-core/esm2015/lib/WebPivotUISettingsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPivotUISettingsDescription = /*@__PURE__*/ (() => {
    class WebPivotUISettingsDescription extends Description {
        get_type() {
            return "WebPivotUISettings";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = false;
            this.i = false;
            this.m = null;
            this.l = null;
        }
        get showConfiguration() {
            return this.h;
        }
        set showConfiguration(a) {
            this.h = a;
            this.g("ShowConfiguration");
        }
        get showRowHeaders() {
            return this.i;
        }
        set showRowHeaders(a) {
            this.i = a;
            this.g("ShowRowHeaders");
        }
        get rowLayout() {
            return this.m;
        }
        set rowLayout(a) {
            this.m = a;
            this.g("RowLayout");
        }
        get horizontalSummariesPosition() {
            return this.l;
        }
        set horizontalSummariesPosition(a) {
            this.l = a;
            this.g("HorizontalSummariesPosition");
        }
    }
    WebPivotUISettingsDescription.$t = /*@__PURE__*/ markType(WebPivotUISettingsDescription, 'WebPivotUISettingsDescription', Description.$);
    return WebPivotUISettingsDescription;
})();
