// igniteui-react-core/esm2015/lib/WebProgressBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebProgressBaseDescription = /*@__PURE__*/ (() => {
    class WebProgressBaseDescription extends Description {
        get_type() {
            return "WebProgressBase";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.m = 0;
            this.n = 0;
            this.s = null;
            this.l = 0;
            this.i = false;
            this.h = false;
            this.r = null;
        }
        get max() {
            return this.m;
        }
        set max(a) {
            this.m = a;
            this.g("Max");
        }
        get value() {
            return this.n;
        }
        set value(a) {
            this.n = a;
            this.g("Value");
        }
        get variant() {
            return this.s;
        }
        set variant(a) {
            this.s = a;
            this.g("Variant");
        }
        get animationDuration() {
            return this.l;
        }
        set animationDuration(a) {
            this.l = a;
            this.g("AnimationDuration");
        }
        get indeterminate() {
            return this.i;
        }
        set indeterminate(a) {
            this.i = a;
            this.g("Indeterminate");
        }
        get hideLabel() {
            return this.h;
        }
        set hideLabel(a) {
            this.h = a;
            this.g("HideLabel");
        }
        get labelFormat() {
            return this.r;
        }
        set labelFormat(a) {
            this.r = a;
            this.g("LabelFormat");
        }
    }
    WebProgressBaseDescription.$t = /*@__PURE__*/ markType(WebProgressBaseDescription, 'WebProgressBaseDescription', Description.$);
    return WebProgressBaseDescription;
})();
