// igniteui-react-core/esm2015/lib/WebPivotKeysDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebPivotKeysDescription = /*@__PURE__*/ (() => {
    class WebPivotKeysDescription extends Description {
        get_type() {
            return "WebPivotKeys";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = null;
            this.n = null;
            this.j = null;
            this.m = null;
            this.l = null;
            this.o = null;
        }
        get children() {
            return this.k;
        }
        set children(a) {
            this.k = a;
            this.g("Children");
        }
        get records() {
            return this.n;
        }
        set records(a) {
            this.n = a;
            this.g("Records");
        }
        get aggregations() {
            return this.j;
        }
        set aggregations(a) {
            this.j = a;
            this.g("Aggregations");
        }
        get level() {
            return this.m;
        }
        set level(a) {
            this.m = a;
            this.g("Level");
        }
        get columnDimensionSeparator() {
            return this.l;
        }
        set columnDimensionSeparator(a) {
            this.l = a;
            this.g("ColumnDimensionSeparator");
        }
        get rowDimensionSeparator() {
            return this.o;
        }
        set rowDimensionSeparator(a) {
            this.o = a;
            this.g("RowDimensionSeparator");
        }
    }
    WebPivotKeysDescription.$t = /*@__PURE__*/ markType(WebPivotKeysDescription, 'WebPivotKeysDescription', Description.$);
    WebPivotKeysDescription.__marshalByValue = true;
    WebPivotKeysDescription.__marshalByValueAlias = "PivotKeys";
    return WebPivotKeysDescription;
})();
