// igniteui-react-core/esm2015/lib/WebRowSelectorTemplateDetailsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowSelectorTemplateDetailsDescription = /*@__PURE__*/ (() => {
    class WebRowSelectorTemplateDetailsDescription extends Description {
        get_type() {
            return "WebRowSelectorTemplateDetails";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = 0;
            this.m = null;
            this.l = null;
            this.h = false;
        }
        get index() {
            return this.j;
        }
        set index(a) {
            this.j = a;
            this.g("Index");
        }
        get rowID() {
            return this.m;
        }
        set rowID(a) {
            this.m = a;
            this.g("RowID");
        }
        get key() {
            return this.l;
        }
        set key(a) {
            this.l = a;
            this.g("Key");
        }
        get selected() {
            return this.h;
        }
        set selected(a) {
            this.h = a;
            this.g("Selected");
        }
    }
    WebRowSelectorTemplateDetailsDescription.$t = /*@__PURE__*/ markType(WebRowSelectorTemplateDetailsDescription, 'WebRowSelectorTemplateDetailsDescription', Description.$);
    return WebRowSelectorTemplateDetailsDescription;
})();
