// igniteui-react-core/esm2015/lib/WebRowDataCancelableEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowDataCancelableEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebRowDataCancelableEventArgsDetailDescription extends Description {
        get_type() {
            return "WebRowDataCancelableEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.j = null;
            this.q = null;
            this.p = null;
            this.m = false;
            this.h = null;
            this.o = null;
            this.s = null;
            this.r = null;
            this.t = null;
        }
        get cellID() {
            return this.j;
        }
        set cellID(a) {
            this.j = a;
            this.g("CellID");
        }
        get oldValue() {
            return this.q;
        }
        set oldValue(a) {
            this.q = a;
            this.g("OldValue");
        }
        get newValue() {
            return this.p;
        }
        set newValue(a) {
            this.p = a;
            this.g("NewValue");
        }
        get isAddRow() {
            return this.m;
        }
        set isAddRow(a) {
            this.m = a;
            this.g("IsAddRow");
        }
        get owner() {
            return this.h;
        }
        set owner(a) {
            this.h = a;
            this.g("Owner");
        }
        get data() {
            return this.o;
        }
        set data(a) {
            this.o = a;
            this.g("Data");
        }
        get rowData() {
            return this.s;
        }
        set rowData(a) {
            this.s = a;
            this.g("RowData");
        }
        get primaryKey() {
            return this.r;
        }
        set primaryKey(a) {
            this.r = a;
            this.g("PrimaryKey");
        }
        get rowKey() {
            return this.t;
        }
        set rowKey(a) {
            this.t = a;
            this.g("RowKey");
        }
    }
    WebRowDataCancelableEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebRowDataCancelableEventArgsDetailDescription, 'WebRowDataCancelableEventArgsDetailDescription', Description.$);
    WebRowDataCancelableEventArgsDetailDescription.__marshalByValue = true;
    WebRowDataCancelableEventArgsDetailDescription.__marshalByValueAlias = "RowDataCancelableEventArgsDetail";
    return WebRowDataCancelableEventArgsDetailDescription;
})();
