// igniteui-react-core/esm2015/lib/WebRowDirectiveDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowDirectiveDescription = /*@__PURE__*/ (() => {
    class WebRowDirectiveDescription extends Description {
        get_type() {
            return "WebRowDirective";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.p = null;
            this.n = 0;
            this.h = false;
            this.j = false;
            this.i = false;
        }
        get data() {
            return this.p;
        }
        set data(a) {
            this.p = a;
            this.g("Data");
        }
        get index() {
            return this.n;
        }
        set index(a) {
            this.n = a;
            this.g("Index");
        }
        get disabled() {
            return this.h;
        }
        set disabled(a) {
            this.h = a;
            this.g("Disabled");
        }
        get pinned() {
            return this.j;
        }
        set pinned(a) {
            this.j = a;
            this.g("Pinned");
        }
        get expanded() {
            return this.i;
        }
        set expanded(a) {
            this.i = a;
            this.g("Expanded");
        }
    }
    WebRowDirectiveDescription.$t = /*@__PURE__*/ markType(WebRowDirectiveDescription, 'WebRowDirectiveDescription', Description.$);
    return WebRowDirectiveDescription;
})();
