// igniteui-react-core/esm2015/lib/WebRowExportingEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowExportingEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebRowExportingEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebRowExportingEventArgsDetail";
        }
        constructor() {
            super();
            this.r = null;
            this.p = 0;
            this.n = false;
        }
        get rowData() {
            return this.r;
        }
        set rowData(a) {
            this.r = a;
            this.g("RowData");
        }
        get rowIndex() {
            return this.p;
        }
        set rowIndex(a) {
            this.p = a;
            this.g("RowIndex");
        }
        get cancel() {
            return this.n;
        }
        set cancel(a) {
            this.n = a;
            this.g("Cancel");
        }
    }
    WebRowExportingEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebRowExportingEventArgsDetailDescription, 'WebRowExportingEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebRowExportingEventArgsDetailDescription.__marshalByValue1 = true;
    WebRowExportingEventArgsDetailDescription.__marshalByValueAlias1 = "RowExportingEventArgsDetail";
    return WebRowExportingEventArgsDetailDescription;
})();
