// igniteui-react-core/esm2015/lib/WebRatingDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRatingDescription = /*@__PURE__*/ (() => {
    class WebRatingDescription extends Description {
        get_type() {
            return "WebRating";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.t = 0;
            this.u = 0;
            this.ad = null;
            this.af = null;
            this.v = 0;
            this.j = false;
            this.l = false;
            this.m = false;
            this.h = false;
            this.i = false;
            this.z = null;
            this.ae = null;
            this.k = false;
            this.ab = null;
            this.ac = null;
        }
        get max() {
            return this.t;
        }
        set max(a) {
            this.t = a;
            this.g("Max");
        }
        get step() {
            return this.u;
        }
        set step(a) {
            this.u = a;
            this.g("Step");
        }
        get label() {
            return this.ad;
        }
        set label(a) {
            this.ad = a;
            this.g("Label");
        }
        get valueFormat() {
            return this.af;
        }
        set valueFormat(a) {
            this.af = a;
            this.g("ValueFormat");
        }
        get value() {
            return this.v;
        }
        set value(a) {
            this.v = a;
            this.g("Value");
        }
        get hoverPreview() {
            return this.j;
        }
        set hoverPreview(a) {
            this.j = a;
            this.g("HoverPreview");
        }
        get readOnly() {
            return this.l;
        }
        set readOnly(a) {
            this.l = a;
            this.g("ReadOnly");
        }
        get single() {
            return this.m;
        }
        set single(a) {
            this.m = a;
            this.g("Single");
        }
        get allowReset() {
            return this.h;
        }
        set allowReset(a) {
            this.h = a;
            this.g("AllowReset");
        }
        get disabled() {
            return this.i;
        }
        set disabled(a) {
            this.i = a;
            this.g("Disabled");
        }
        get defaultValue() {
            return this.z;
        }
        set defaultValue(a) {
            this.z = a;
            this.g("DefaultValue");
        }
        get name() {
            return this.ae;
        }
        set name(a) {
            this.ae = a;
            this.g("Name");
        }
        get invalid() {
            return this.k;
        }
        set invalid(a) {
            this.k = a;
            this.g("Invalid");
        }
        get changeRef() {
            return this.ab;
        }
        set changeRef(a) {
            this.ab = a;
            this.g("ChangeRef");
        }
        get hoverRef() {
            return this.ac;
        }
        set hoverRef(a) {
            this.ac = a;
            this.g("HoverRef");
        }
    }
    WebRatingDescription.$t = /*@__PURE__*/ markType(WebRatingDescription, 'WebRatingDescription', Description.$);
    return WebRatingDescription;
})();
