// igniteui-react-core/esm2015/lib/WebRangeSliderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebSliderBaseDescription } from "./WebSliderBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRangeSliderDescription = /*@__PURE__*/ (() => {
    class WebRangeSliderDescription extends WebSliderBaseDescription {
        get_type() {
            return "WebRangeSlider";
        }
        constructor() {
            super();
            this.aq = 0;
            this.ar = 0;
            this.aw = null;
            this.ax = null;
            this.av = null;
            this.au = null;
        }
        get lower() {
            return this.aq;
        }
        set lower(a) {
            this.aq = a;
            this.g("Lower");
        }
        get upper() {
            return this.ar;
        }
        set upper(a) {
            this.ar = a;
            this.g("Upper");
        }
        get thumbLabelLower() {
            return this.aw;
        }
        set thumbLabelLower(a) {
            this.aw = a;
            this.g("ThumbLabelLower");
        }
        get thumbLabelUpper() {
            return this.ax;
        }
        set thumbLabelUpper(a) {
            this.ax = a;
            this.g("ThumbLabelUpper");
        }
        get inputRef() {
            return this.av;
        }
        set inputRef(a) {
            this.av = a;
            this.g("InputRef");
        }
        get changeRef() {
            return this.au;
        }
        set changeRef(a) {
            this.au = a;
            this.g("ChangeRef");
        }
    }
    WebRangeSliderDescription.$t = /*@__PURE__*/ markType(WebRangeSliderDescription, 'WebRangeSliderDescription', WebSliderBaseDescription.$);
    return WebRangeSliderDescription;
})();
