// igniteui-react-core/esm2015/lib/WebRowToggleEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowToggleEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebRowToggleEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        get_type() {
            return "WebRowToggleEventArgsDetail";
        }
        constructor() {
            super();
            this.r = null;
            this.s = null;
            this.o = false;
            this.n = false;
        }
        get rowID() {
            return this.r;
        }
        set rowID(a) {
            this.r = a;
            this.g("RowID");
        }
        get rowKey() {
            return this.s;
        }
        set rowKey(a) {
            this.s = a;
            this.g("RowKey");
        }
        get expanded() {
            return this.o;
        }
        set expanded(a) {
            this.o = a;
            this.g("Expanded");
        }
        get cancel() {
            return this.n;
        }
        set cancel(a) {
            this.n = a;
            this.g("Cancel");
        }
    }
    WebRowToggleEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebRowToggleEventArgsDetailDescription, 'WebRowToggleEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebRowToggleEventArgsDetailDescription.__marshalByValue1 = true;
    WebRowToggleEventArgsDetailDescription.__marshalByValueAlias1 = "RowToggleEventArgsDetail";
    return WebRowToggleEventArgsDetailDescription;
})();
