// igniteui-react-core/esm2015/lib/WebRowDragStartEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowDragStartEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebRowDragStartEventArgsDetailDescription extends Description {
        get_type() {
            return "WebRowDragStartEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.m = null;
            this.h = null;
            this.k = false;
            this.n = null;
        }
        get dragDirective() {
            return this.m;
        }
        set dragDirective(a) {
            this.m = a;
            this.g("DragDirective");
        }
        get dragData() {
            return this.h;
        }
        set dragData(a) {
            this.h = a;
            this.g("DragData");
        }
        get cancel() {
            return this.k;
        }
        set cancel(a) {
            this.k = a;
            this.g("Cancel");
        }
        get owner() {
            return this.n;
        }
        set owner(a) {
            this.n = a;
            this.g("Owner");
        }
    }
    WebRowDragStartEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebRowDragStartEventArgsDetailDescription, 'WebRowDragStartEventArgsDetailDescription', Description.$);
    WebRowDragStartEventArgsDetailDescription.__marshalByValue = true;
    WebRowDragStartEventArgsDetailDescription.__marshalByValueAlias = "RowDragStartEventArgsDetail";
    return WebRowDragStartEventArgsDetailDescription;
})();
