// igniteui-react-core/esm2015/lib/WebSortingEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSortingEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebSortingEventArgsDetailDescription extends Description {
        get_type() {
            return "WebSortingEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
            this.o = null;
            this.m = false;
        }
        get sortingExpressions() {
            return this.i;
        }
        set sortingExpressions(a) {
            this.i = a;
            this.g("SortingExpressions");
        }
        get groupingExpressions() {
            return this.h;
        }
        set groupingExpressions(a) {
            this.h = a;
            this.g("GroupingExpressions");
        }
        get owner() {
            return this.o;
        }
        set owner(a) {
            this.o = a;
            this.g("Owner");
        }
        get cancel() {
            return this.m;
        }
        set cancel(a) {
            this.m = a;
            this.g("Cancel");
        }
    }
    WebSortingEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebSortingEventArgsDetailDescription, 'WebSortingEventArgsDetailDescription', Description.$);
    WebSortingEventArgsDetailDescription.__marshalByValue = true;
    WebSortingEventArgsDetailDescription.__marshalByValueAlias = "SortingEventArgsDetail";
    return WebSortingEventArgsDetailDescription;
})();
