// igniteui-react-core/esm2015/lib/WebSearchInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseSearchInfoDescription } from "./WebBaseSearchInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSearchInfoDescription = /*@__PURE__*/ (() => {
    class WebSearchInfoDescription extends WebBaseSearchInfoDescription {
        get_type() {
            return "WebSearchInfo";
        }
        constructor() {
            super();
            this.u = null;
            this.s = 0;
        }
        get matchInfoCacheRef() {
            return this.u;
        }
        set matchInfoCacheRef(a) {
            this.u = a;
            this.g("MatchInfoCacheRef");
        }
        get activeMatchIndex() {
            return this.s;
        }
        set activeMatchIndex(a) {
            this.s = a;
            this.g("ActiveMatchIndex");
        }
    }
    WebSearchInfoDescription.$t = /*@__PURE__*/ markType(WebSearchInfoDescription, 'WebSearchInfoDescription', WebBaseSearchInfoDescription.$);
    return WebSearchInfoDescription;
})();
