// igniteui-react-core/esm2015/lib/WebSnackbarDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseAlertLikeDescription } from "./WebBaseAlertLikeDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSnackbarDescription = /*@__PURE__*/ (() => {
    class WebSnackbarDescription extends WebBaseAlertLikeDescription {
        get_type() {
            return "WebSnackbar";
        }
        constructor() {
            super();
            this.r = null;
            this.q = null;
        }
        get actionText() {
            return this.r;
        }
        set actionText(a) {
            this.r = a;
            this.g("ActionText");
        }
        get actionRef() {
            return this.q;
        }
        set actionRef(a) {
            this.q = a;
            this.g("ActionRef");
        }
    }
    WebSnackbarDescription.$t = /*@__PURE__*/ markType(WebSnackbarDescription, 'WebSnackbarDescription', WebBaseAlertLikeDescription.$);
    return WebSnackbarDescription;
})();
