// igniteui-react-core/esm2015/lib/WebSplitPaneDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebDockManagerPaneDescription } from "./WebDockManagerPaneDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSplitPaneDescription = /*@__PURE__*/ (() => {
    class WebSplitPaneDescription extends WebDockManagerPaneDescription {
        get_type() {
            return "WebSplitPane";
        }
        constructor() {
            super();
            this.ae = null;
            this.ag = null;
            this.af = null;
            this.m = null;
            this.z = 0;
            this.o = null;
            this.y = 0;
            this.x = 0;
            this.s = false;
            this.r = false;
            this.t = false;
        }
        get id() {
            return this.ae;
        }
        set id(a) {
            this.ae = a;
            this.g("Id");
        }
        get paneType() {
            return this.ag;
        }
        set paneType(a) {
            this.ag = a;
            this.g("PaneType");
        }
        get orientation() {
            return this.af;
        }
        set orientation(a) {
            this.af = a;
            this.g("Orientation");
        }
        get panes() {
            return this.m;
        }
        set panes(a) {
            this.m = a;
            this.g("Panes");
        }
        get size() {
            return this.z;
        }
        set size(a) {
            this.z = a;
            this.g("Size");
        }
        get floatingLocation() {
            return this.o;
        }
        set floatingLocation(a) {
            this.o = a;
            this.g("FloatingLocation");
        }
        get floatingWidth() {
            return this.y;
        }
        set floatingWidth(a) {
            this.y = a;
            this.g("FloatingWidth");
        }
        get floatingHeight() {
            return this.x;
        }
        set floatingHeight(a) {
            this.x = a;
            this.g("FloatingHeight");
        }
        get floatingResizable() {
            return this.s;
        }
        set floatingResizable(a) {
            this.s = a;
            this.g("FloatingResizable");
        }
        get allowEmpty() {
            return this.r;
        }
        set allowEmpty(a) {
            this.r = a;
            this.g("AllowEmpty");
        }
        get isMaximized() {
            return this.t;
        }
        set isMaximized(a) {
            this.t = a;
            this.g("IsMaximized");
        }
    }
    WebSplitPaneDescription.$t = /*@__PURE__*/ markType(WebSplitPaneDescription, 'WebSplitPaneDescription', WebDockManagerPaneDescription.$);
    WebSplitPaneDescription.__marshalByValue1 = true;
    WebSplitPaneDescription.__marshalByValueAlias1 = "SplitPane";
    return WebSplitPaneDescription;
})();
