// igniteui-react-core/esm2015/lib/WebSortingExpressionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDescription } from "./WebBaseEventArgsDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSortingExpressionDescription = /*@__PURE__*/ (() => {
    class WebSortingExpressionDescription extends WebBaseEventArgsDescription {
        get_type() {
            return "WebSortingExpression";
        }
        constructor() {
            super();
            this.t = null;
            this.s = null;
            this.p = false;
            this.m = null;
        }
        get fieldName() {
            return this.t;
        }
        set fieldName(a) {
            this.t = a;
            this.g("FieldName");
        }
        get dir() {
            return this.s;
        }
        set dir(a) {
            this.s = a;
            this.g("Dir");
        }
        get ignoreCase() {
            return this.p;
        }
        set ignoreCase(a) {
            this.p = a;
            this.g("IgnoreCase");
        }
        get strategy() {
            return this.m;
        }
        set strategy(a) {
            this.m = a;
            this.g("Strategy");
        }
    }
    WebSortingExpressionDescription.$t = /*@__PURE__*/ markType(WebSortingExpressionDescription, 'WebSortingExpressionDescription', WebBaseEventArgsDescription.$);
    WebSortingExpressionDescription.__marshalByValue1 = true;
    WebSortingExpressionDescription.__marshalByValueAlias1 = "SortingExpression";
    return WebSortingExpressionDescription;
})();
