// igniteui-react-core/esm2015/lib/WebStepDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebStepDescription = /*@__PURE__*/ (() => {
    class WebStepDescription extends Description {
        get_type() {
            return "WebStep";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.k = false;
            this.h = false;
            this.l = false;
            this.j = false;
            this.i = false;
        }
        get invalid() {
            return this.k;
        }
        set invalid(a) {
            this.k = a;
            this.g("Invalid");
        }
        get active() {
            return this.h;
        }
        set active(a) {
            this.h = a;
            this.g("Active");
        }
        get optional() {
            return this.l;
        }
        set optional(a) {
            this.l = a;
            this.g("Optional");
        }
        get disabled() {
            return this.j;
        }
        set disabled(a) {
            this.j = a;
            this.g("Disabled");
        }
        get complete() {
            return this.i;
        }
        set complete(a) {
            this.i = a;
            this.g("Complete");
        }
    }
    WebStepDescription.$t = /*@__PURE__*/ markType(WebStepDescription, 'WebStepDescription', Description.$);
    return WebStepDescription;
})();
