// igniteui-react-core/esm2015/lib/WebSummaryResultDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSummaryResultDescription = /*@__PURE__*/ (() => {
    class WebSummaryResultDescription extends Description {
        get_type() {
            return "WebSummaryResult";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.n = null;
            this.o = null;
            this.k = null;
            this.i = false;
        }
        get key() {
            return this.n;
        }
        set key(a) {
            this.n = a;
            this.g("Key");
        }
        get label() {
            return this.o;
        }
        set label(a) {
            this.o = a;
            this.g("Label");
        }
        get result() {
            return this.k;
        }
        set result(a) {
            this.k = a;
            this.g("Result");
        }
        get defaultFormatting() {
            return this.i;
        }
        set defaultFormatting(a) {
            this.i = a;
            this.g("DefaultFormatting");
        }
    }
    WebSummaryResultDescription.$t = /*@__PURE__*/ markType(WebSummaryResultDescription, 'WebSummaryResultDescription', Description.$);
    WebSummaryResultDescription.__marshalByValue = true;
    WebSummaryResultDescription.__marshalByValueAlias = "SummaryResult";
    return WebSummaryResultDescription;
})();
