// igniteui-react-core/esm2015/lib/WebToggleButtonDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebToggleButtonDescription = /*@__PURE__*/ (() => {
    class WebToggleButtonDescription extends Description {
        get_type() {
            return "WebToggleButton";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.l = null;
            this.i = false;
            this.h = false;
        }
        get value() {
            return this.l;
        }
        set value(a) {
            this.l = a;
            this.g("Value");
        }
        get selected() {
            return this.i;
        }
        set selected(a) {
            this.i = a;
            this.g("Selected");
        }
        get disabled() {
            return this.h;
        }
        set disabled(a) {
            this.h = a;
            this.g("Disabled");
        }
    }
    WebToggleButtonDescription.$t = /*@__PURE__*/ markType(WebToggleButtonDescription, 'WebToggleButtonDescription', Description.$);
    return WebToggleButtonDescription;
})();
