// igniteui-react-core/esm2015/lib/WebTextareaDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebTextareaDescription = /*@__PURE__*/ (() => {
    class WebTextareaDescription extends Description {
        get_type() {
            return "WebTextarea";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.ae = null;
            this.ad = null;
            this.aj = null;
            this.ak = null;
            this.v = 0;
            this.w = 0;
            this.j = false;
            this.am = null;
            this.k = false;
            this.an = null;
            this.x = 0;
            this.ao = null;
            this.m = false;
            this.ap = null;
            this.n = false;
            this.h = false;
            this.l = false;
            this.ab = null;
            this.al = null;
            this.i = false;
            this.ai = null;
            this.ag = null;
            this.ah = null;
            this.af = null;
        }
        get autocomplete() {
            return this.ae;
        }
        set autocomplete(a) {
            this.ae = a;
            this.g("Autocomplete");
        }
        get autocapitalize() {
            return this.ad;
        }
        set autocapitalize(a) {
            this.ad = a;
            this.g("Autocapitalize");
        }
        get inputMode() {
            return this.aj;
        }
        set inputMode(a) {
            this.aj = a;
            this.g("InputMode");
        }
        get label() {
            return this.ak;
        }
        set label(a) {
            this.ak = a;
            this.g("Label");
        }
        get maxLength() {
            return this.v;
        }
        set maxLength(a) {
            this.v = a;
            this.g("MaxLength");
        }
        get minLength() {
            return this.w;
        }
        set minLength(a) {
            this.w = a;
            this.g("MinLength");
        }
        get outlined() {
            return this.j;
        }
        set outlined(a) {
            this.j = a;
            this.g("Outlined");
        }
        get placeholder() {
            return this.am;
        }
        set placeholder(a) {
            this.am = a;
            this.g("Placeholder");
        }
        get readOnly() {
            return this.k;
        }
        set readOnly(a) {
            this.k = a;
            this.g("ReadOnly");
        }
        get resize() {
            return this.an;
        }
        set resize(a) {
            this.an = a;
            this.g("Resize");
        }
        get rows() {
            return this.x;
        }
        set rows(a) {
            this.x = a;
            this.g("Rows");
        }
        get value() {
            return this.ao;
        }
        set value(a) {
            this.ao = a;
            this.g("Value");
        }
        get spellcheck() {
            return this.m;
        }
        set spellcheck(a) {
            this.m = a;
            this.g("Spellcheck");
        }
        get wrap() {
            return this.ap;
        }
        set wrap(a) {
            this.ap = a;
            this.g("Wrap");
        }
        get validateOnly() {
            return this.n;
        }
        set validateOnly(a) {
            this.n = a;
            this.g("ValidateOnly");
        }
        get disabled() {
            return this.h;
        }
        set disabled(a) {
            this.h = a;
            this.g("Disabled");
        }
        get required() {
            return this.l;
        }
        set required(a) {
            this.l = a;
            this.g("Required");
        }
        get defaultValue() {
            return this.ab;
        }
        set defaultValue(a) {
            this.ab = a;
            this.g("DefaultValue");
        }
        get name() {
            return this.al;
        }
        set name(a) {
            this.al = a;
            this.g("Name");
        }
        get invalid() {
            return this.i;
        }
        set invalid(a) {
            this.i = a;
            this.g("Invalid");
        }
        get inputRef() {
            return this.ai;
        }
        set inputRef(a) {
            this.ai = a;
            this.g("InputRef");
        }
        get changeRef() {
            return this.ag;
        }
        set changeRef(a) {
            this.ag = a;
            this.g("ChangeRef");
        }
        get focusRef() {
            return this.ah;
        }
        set focusRef(a) {
            this.ah = a;
            this.g("FocusRef");
        }
        get blurRef() {
            return this.af;
        }
        set blurRef(a) {
            this.af = a;
            this.g("BlurRef");
        }
    }
    WebTextareaDescription.$t = /*@__PURE__*/ markType(WebTextareaDescription, 'WebTextareaDescription', Description.$);
    return WebTextareaDescription;
})();
