// igniteui-react-core/esm2015/lib/WebTreeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebTreeDescription = /*@__PURE__*/ (() => {
    class WebTreeDescription extends Description {
        get_type() {
            return "WebTree";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = false;
            this.i = false;
            this.q = null;
            this.r = null;
            this.p = null;
            this.o = null;
            this.n = null;
            this.m = null;
            this.l = null;
        }
        get singleBranchExpand() {
            return this.h;
        }
        set singleBranchExpand(a) {
            this.h = a;
            this.g("SingleBranchExpand");
        }
        get toggleNodeOnClick() {
            return this.i;
        }
        set toggleNodeOnClick(a) {
            this.i = a;
            this.g("ToggleNodeOnClick");
        }
        get selection() {
            return this.q;
        }
        set selection(a) {
            this.q = a;
            this.g("Selection");
        }
        get selectionChangedRef() {
            return this.r;
        }
        set selectionChangedRef(a) {
            this.r = a;
            this.g("SelectionChangedRef");
        }
        get itemExpandingRef() {
            return this.p;
        }
        set itemExpandingRef(a) {
            this.p = a;
            this.g("ItemExpandingRef");
        }
        get itemExpandedRef() {
            return this.o;
        }
        set itemExpandedRef(a) {
            this.o = a;
            this.g("ItemExpandedRef");
        }
        get itemCollapsingRef() {
            return this.n;
        }
        set itemCollapsingRef(a) {
            this.n = a;
            this.g("ItemCollapsingRef");
        }
        get itemCollapsedRef() {
            return this.m;
        }
        set itemCollapsedRef(a) {
            this.m = a;
            this.g("ItemCollapsedRef");
        }
        get activeItemRef() {
            return this.l;
        }
        set activeItemRef(a) {
            this.l = a;
            this.g("ActiveItemRef");
        }
    }
    WebTreeDescription.$t = /*@__PURE__*/ markType(WebTreeDescription, 'WebTreeDescription', Description.$);
    return WebTreeDescription;
})();
