// igniteui-react-core/esm2015/lib/WebTreeGridRecordDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebTreeGridRecordDescription = /*@__PURE__*/ (() => {
    class WebTreeGridRecordDescription extends Description {
        get_type() {
            return "WebTreeGridRecord";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.t = null;
            this.s = null;
            this.h = null;
            this.j = null;
            this.q = 0;
            this.n = false;
            this.m = false;
        }
        get key() {
            return this.t;
        }
        set key(a) {
            this.t = a;
            this.g("Key");
        }
        get data() {
            return this.s;
        }
        set data(a) {
            this.s = a;
            this.g("Data");
        }
        get children() {
            return this.h;
        }
        set children(a) {
            this.h = a;
            this.g("Children");
        }
        get recordParent() {
            return this.j;
        }
        set recordParent(a) {
            this.j = a;
            this.g("RecordParent");
        }
        get level() {
            return this.q;
        }
        set level(a) {
            this.q = a;
            this.g("Level");
        }
        get isFilteredOutParent() {
            return this.n;
        }
        set isFilteredOutParent(a) {
            this.n = a;
            this.g("IsFilteredOutParent");
        }
        get expanded() {
            return this.m;
        }
        set expanded(a) {
            this.m = a;
            this.g("Expanded");
        }
    }
    WebTreeGridRecordDescription.$t = /*@__PURE__*/ markType(WebTreeGridRecordDescription, 'WebTreeGridRecordDescription', Description.$);
    WebTreeGridRecordDescription.__marshalByValue = true;
    WebTreeGridRecordDescription.__marshalByValueAlias = "TreeGridRecord";
    return WebTreeGridRecordDescription;
})();
