// igniteui-react-core/esm2015/lib/WebTreeItemDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebTreeItemDescription = /*@__PURE__*/ (() => {
    class WebTreeItemDescription extends Description {
        get_type() {
            return "WebTreeItem";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = null;
            this.t = 0;
            this.x = null;
            this.l = false;
            this.j = false;
            this.k = false;
            this.n = false;
            this.m = false;
            this.v = null;
        }
        get parent() {
            return this.h;
        }
        set parent(a) {
            this.h = a;
            this.g("Parent");
        }
        get level() {
            return this.t;
        }
        set level(a) {
            this.t = a;
            this.g("Level");
        }
        get label() {
            return this.x;
        }
        set label(a) {
            this.x = a;
            this.g("Label");
        }
        get expanded() {
            return this.l;
        }
        set expanded(a) {
            this.l = a;
            this.g("Expanded");
        }
        get active() {
            return this.j;
        }
        set active(a) {
            this.j = a;
            this.g("Active");
        }
        get disabled() {
            return this.k;
        }
        set disabled(a) {
            this.k = a;
            this.g("Disabled");
        }
        get selected() {
            return this.n;
        }
        set selected(a) {
            this.n = a;
            this.g("Selected");
        }
        get loading() {
            return this.m;
        }
        set loading(a) {
            this.m = a;
            this.g("Loading");
        }
        get value() {
            return this.v;
        }
        set value(a) {
            this.v = a;
            this.g("Value");
        }
    }
    WebTreeItemDescription.$t = /*@__PURE__*/ markType(WebTreeItemDescription, 'WebTreeItemDescription', Description.$);
    return WebTreeItemDescription;
})();
