// igniteui-react-core/esm2015/lib/WebTabsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebTabsDescription = /*@__PURE__*/ (() => {
    class WebTabsDescription extends Description {
        get_type() {
            return "WebTabs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.i = null;
            this.h = null;
            this.j = null;
        }
        get alignment() {
            return this.i;
        }
        set alignment(a) {
            this.i = a;
            this.g("Alignment");
        }
        get activation() {
            return this.h;
        }
        set activation(a) {
            this.h = a;
            this.g("Activation");
        }
        get changeRef() {
            return this.j;
        }
        set changeRef(a) {
            this.j = a;
            this.g("ChangeRef");
        }
    }
    WebTabsDescription.$t = /*@__PURE__*/ markType(WebTabsDescription, 'WebTabsDescription', Description.$);
    return WebTabsDescription;
})();
