// igniteui-react-core/esm2015/lib/XBaseDataType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "./type";
/**
 * Enum for XBase field data types.
 */
export var XBaseDataType = /*@__PURE__*/ (function (XBaseDataType) {
    /**
     * Character data type.
     */
    XBaseDataType[XBaseDataType["Character"] = 0] = "Character";
    /**
     * Number data type.
     */
    XBaseDataType[XBaseDataType["Number"] = 1] = "Number";
    /**
     * Logical data type.
     */
    XBaseDataType[XBaseDataType["Logical"] = 2] = "Logical";
    /**
     * Date data type.
     */
    XBaseDataType[XBaseDataType["Date"] = 3] = "Date";
    /**
     * Memo data type.
     */
    XBaseDataType[XBaseDataType["Memo"] = 4] = "Memo";
    /**
     * Floating point data type.
     */
    XBaseDataType[XBaseDataType["FloatingPoint"] = 5] = "FloatingPoint";
    /**
     * Binary data type.
     */
    XBaseDataType[XBaseDataType["Binary"] = 6] = "Binary";
    /**
     * General data type.
     */
    XBaseDataType[XBaseDataType["General"] = 7] = "General";
    /**
     * Picture data type.
     */
    XBaseDataType[XBaseDataType["Picture"] = 8] = "Picture";
    /**
     * Currency data type.
     */
    XBaseDataType[XBaseDataType["Currency"] = 9] = "Currency";
    /**
     * DateTime data type.
     */
    XBaseDataType[XBaseDataType["DateTime"] = 10] = "DateTime";
    /**
     * Integer data type.
     */
    XBaseDataType[XBaseDataType["Integer"] = 11] = "Integer";
    /**
     * VariField data type.
     */
    XBaseDataType[XBaseDataType["VariField"] = 12] = "VariField";
    /**
     * Variant data type.
     */
    XBaseDataType[XBaseDataType["Variant"] = 13] = "Variant";
    /**
     * Timestamp data type.
     */
    XBaseDataType[XBaseDataType["Timestamp"] = 14] = "Timestamp";
    /**
     * Double data type.
     */
    XBaseDataType[XBaseDataType["double1"] = 15] = "double1";
    /**
     * Auto increment data type.
     */
    XBaseDataType[XBaseDataType["AutoIncrement"] = 16] = "AutoIncrement";
    return XBaseDataType;
})({});
/**
 * @hidden
 */
export let XBaseDataType_$type = /*@__PURE__*/ markEnum('XBaseDataType', 'Character,0|Number,1|Logical,2|Date,3|Memo,4|FloatingPoint,5|Binary,6|General,7|Picture,8|Currency,9|DateTime,10|Integer,11|VariField,12|Variant,13|Timestamp,14|Double:double1,15|AutoIncrement,16');
