// igniteui-react-core/esm2015/lib/ZoomSliderResolvingAxisValueEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ZoomSliderResolvingAxisValueEventArgsDescription = /*@__PURE__*/ (() => {
    class ZoomSliderResolvingAxisValueEventArgsDescription extends Description {
        get_type() {
            return "ZoomSliderResolvingAxisValueEventArgs";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.h = 0;
            this.j = null;
        }
        get position() {
            return this.h;
        }
        set position(a) {
            this.h = a;
            this.g("Position");
        }
        get valueRef() {
            return this.j;
        }
        set valueRef(a) {
            this.j = a;
            this.g("ValueRef");
        }
    }
    ZoomSliderResolvingAxisValueEventArgsDescription.$t = /*@__PURE__*/ markType(ZoomSliderResolvingAxisValueEventArgsDescription, 'ZoomSliderResolvingAxisValueEventArgsDescription', Description.$);
    return ZoomSliderResolvingAxisValueEventArgsDescription;
})();
