// igniteui-react-core/esm2015/lib/ZoomSliderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ZoomSliderDescription = /*@__PURE__*/ (() => {
    class ZoomSliderDescription extends Description {
        get width() {
            return this.bx;
        }
        set width(a) {
            this.bx = a;
            this.g("Width");
        }
        get height() {
            return this.a8;
        }
        set height(a) {
            this.a8 = a;
            this.g("Height");
        }
        get background() {
            return this.a5;
        }
        set background(a) {
            this.a5 = a;
            this.g("Background");
        }
        get_type() {
            return "ZoomSlider";
        }
        get type() {
            return this.get_type();
        }
        constructor() {
            super();
            this.bx = null;
            this.a8 = null;
            this.a5 = null;
            this.a3 = 0;
            this.z = 0;
            this.ab = 0;
            this.l = 0;
            this.h = null;
            this.aa = 0;
            this.ad = 0;
            this.o = 0;
            this.ag = 0;
            this.af = 0;
            this.m = 0;
            this.bp = null;
            this.bm = null;
            this.x = 0;
            this.s = 0;
            this.be = null;
            this.bn = null;
            this.bf = null;
            this.bo = null;
            this.bg = null;
            this.y = 0;
            this.t = 0;
            this.w = 0;
            this.r = 0;
            this.bk = null;
            this.bl = null;
            this.v = 0;
            this.bc = null;
            this.bd = null;
            this.q = 0;
            this.a6 = null;
            this.a7 = null;
            this.n = 0;
            this.bq = null;
            this.br = null;
            this.ac = 0;
            this.bs = null;
            this.bh = null;
            this.bj = null;
            this.bi = null;
            this.u = 0;
            this.a9 = null;
            this.bb = null;
            this.ba = null;
            this.p = 0;
            this.j = false;
            this.bu = null;
            this.ae = 0;
            this.bv = null;
            this.bw = null;
            this.bt = null;
            this.by = null;
        }
        get panTransitionDuration() {
            return this.a3;
        }
        set panTransitionDuration(a) {
            this.a3 = a;
            this.g("PanTransitionDuration");
        }
        get maxZoomWidth() {
            return this.z;
        }
        set maxZoomWidth(a) {
            this.z = a;
            this.g("MaxZoomWidth");
        }
        get pixelScalingRatio() {
            return this.ab;
        }
        set pixelScalingRatio(a) {
            this.ab = a;
            this.g("PixelScalingRatio");
        }
        get actualPixelScalingRatio() {
            return this.l;
        }
        set actualPixelScalingRatio(a) {
            this.l = a;
            this.g("ActualPixelScalingRatio");
        }
        get windowRect() {
            return this.h;
        }
        set windowRect(a) {
            this.h = a;
            this.g("WindowRect");
        }
        get minZoomWidth() {
            return this.aa;
        }
        set minZoomWidth(a) {
            this.aa = a;
            this.g("MinZoomWidth");
        }
        get startInset() {
            return this.ad;
        }
        set startInset(a) {
            this.ad = a;
            this.g("StartInset");
        }
        get endInset() {
            return this.o;
        }
        set endInset(a) {
            this.o = a;
            this.g("EndInset");
        }
        get trackStartInset() {
            return this.ag;
        }
        set trackStartInset(a) {
            this.ag = a;
            this.g("TrackStartInset");
        }
        get trackEndInset() {
            return this.af;
        }
        set trackEndInset(a) {
            this.af = a;
            this.g("TrackEndInset");
        }
        get barExtent() {
            return this.m;
        }
        set barExtent(a) {
            this.m = a;
            this.g("BarExtent");
        }
        get orientation() {
            return this.bp;
        }
        set orientation(a) {
            this.bp = a;
            this.g("Orientation");
        }
        get lowerThumbBrush() {
            return this.bm;
        }
        set lowerThumbBrush(a) {
            this.bm = a;
            this.g("LowerThumbBrush");
        }
        get lowerThumbStrokeThickness() {
            return this.x;
        }
        set lowerThumbStrokeThickness(a) {
            this.x = a;
            this.g("LowerThumbStrokeThickness");
        }
        get higherThumbStrokeThickness() {
            return this.s;
        }
        set higherThumbStrokeThickness(a) {
            this.s = a;
            this.g("HigherThumbStrokeThickness");
        }
        get higherThumbBrush() {
            return this.be;
        }
        set higherThumbBrush(a) {
            this.be = a;
            this.g("HigherThumbBrush");
        }
        get lowerThumbOutline() {
            return this.bn;
        }
        set lowerThumbOutline(a) {
            this.bn = a;
            this.g("LowerThumbOutline");
        }
        get higherThumbOutline() {
            return this.bf;
        }
        set higherThumbOutline(a) {
            this.bf = a;
            this.g("HigherThumbOutline");
        }
        get lowerThumbRidgesBrush() {
            return this.bo;
        }
        set lowerThumbRidgesBrush(a) {
            this.bo = a;
            this.g("LowerThumbRidgesBrush");
        }
        get higherThumbRidgesBrush() {
            return this.bg;
        }
        set higherThumbRidgesBrush(a) {
            this.bg = a;
            this.g("HigherThumbRidgesBrush");
        }
        get lowerThumbWidth() {
            return this.y;
        }
        set lowerThumbWidth(a) {
            this.y = a;
            this.g("LowerThumbWidth");
        }
        get higherThumbWidth() {
            return this.t;
        }
        set higherThumbWidth(a) {
            this.t = a;
            this.g("HigherThumbWidth");
        }
        get lowerThumbHeight() {
            return this.w;
        }
        set lowerThumbHeight(a) {
            this.w = a;
            this.g("LowerThumbHeight");
        }
        get higherThumbHeight() {
            return this.r;
        }
        set higherThumbHeight(a) {
            this.r = a;
            this.g("HigherThumbHeight");
        }
        get lowerShadeBrush() {
            return this.bk;
        }
        set lowerShadeBrush(a) {
            this.bk = a;
            this.g("LowerShadeBrush");
        }
        get lowerShadeOutline() {
            return this.bl;
        }
        set lowerShadeOutline(a) {
            this.bl = a;
            this.g("LowerShadeOutline");
        }
        get lowerShadeStrokeThickness() {
            return this.v;
        }
        set lowerShadeStrokeThickness(a) {
            this.v = a;
            this.g("LowerShadeStrokeThickness");
        }
        get higherShadeBrush() {
            return this.bc;
        }
        set higherShadeBrush(a) {
            this.bc = a;
            this.g("HigherShadeBrush");
        }
        get higherShadeOutline() {
            return this.bd;
        }
        set higherShadeOutline(a) {
            this.bd = a;
            this.g("HigherShadeOutline");
        }
        get higherShadeStrokeThickness() {
            return this.q;
        }
        set higherShadeStrokeThickness(a) {
            this.q = a;
            this.g("HigherShadeStrokeThickness");
        }
        get barBrush() {
            return this.a6;
        }
        set barBrush(a) {
            this.a6 = a;
            this.g("BarBrush");
        }
        get barOutline() {
            return this.a7;
        }
        set barOutline(a) {
            this.a7 = a;
            this.g("BarOutline");
        }
        get barStrokeThickness() {
            return this.n;
        }
        set barStrokeThickness(a) {
            this.n = a;
            this.g("BarStrokeThickness");
        }
        get rangeThumbBrush() {
            return this.bq;
        }
        set rangeThumbBrush(a) {
            this.bq = a;
            this.g("RangeThumbBrush");
        }
        get rangeThumbOutline() {
            return this.br;
        }
        set rangeThumbOutline(a) {
            this.br = a;
            this.g("RangeThumbOutline");
        }
        get rangeThumbStrokeThickness() {
            return this.ac;
        }
        set rangeThumbStrokeThickness(a) {
            this.ac = a;
            this.g("RangeThumbStrokeThickness");
        }
        get rangeThumbRidgesBrush() {
            return this.bs;
        }
        set rangeThumbRidgesBrush(a) {
            this.bs = a;
            this.g("RangeThumbRidgesBrush");
        }
        get lowerCalloutBrush() {
            return this.bh;
        }
        set lowerCalloutBrush(a) {
            this.bh = a;
            this.g("LowerCalloutBrush");
        }
        get lowerCalloutTextColor() {
            return this.bj;
        }
        set lowerCalloutTextColor(a) {
            this.bj = a;
            this.g("LowerCalloutTextColor");
        }
        get lowerCalloutOutline() {
            return this.bi;
        }
        set lowerCalloutOutline(a) {
            this.bi = a;
            this.g("LowerCalloutOutline");
        }
        get lowerCalloutStrokeThickness() {
            return this.u;
        }
        set lowerCalloutStrokeThickness(a) {
            this.u = a;
            this.g("LowerCalloutStrokeThickness");
        }
        get higherCalloutBrush() {
            return this.a9;
        }
        set higherCalloutBrush(a) {
            this.a9 = a;
            this.g("HigherCalloutBrush");
        }
        get higherCalloutTextColor() {
            return this.bb;
        }
        set higherCalloutTextColor(a) {
            this.bb = a;
            this.g("HigherCalloutTextColor");
        }
        get higherCalloutOutline() {
            return this.ba;
        }
        set higherCalloutOutline(a) {
            this.ba = a;
            this.g("HigherCalloutOutline");
        }
        get higherCalloutStrokeThickness() {
            return this.p;
        }
        set higherCalloutStrokeThickness(a) {
            this.p = a;
            this.g("HigherCalloutStrokeThickness");
        }
        get areThumbCalloutsEnabled() {
            return this.j;
        }
        set areThumbCalloutsEnabled(a) {
            this.j = a;
            this.g("AreThumbCalloutsEnabled");
        }
        get thumbCalloutFontFamily() {
            return this.bu;
        }
        set thumbCalloutFontFamily(a) {
            this.bu = a;
            this.g("ThumbCalloutFontFamily");
        }
        get thumbCalloutFontSize() {
            return this.ae;
        }
        set thumbCalloutFontSize(a) {
            this.ae = a;
            this.g("ThumbCalloutFontSize");
        }
        get thumbCalloutFontStyle() {
            return this.bv;
        }
        set thumbCalloutFontStyle(a) {
            this.bv = a;
            this.g("ThumbCalloutFontStyle");
        }
        get thumbCalloutFontWeight() {
            return this.bw;
        }
        set thumbCalloutFontWeight(a) {
            this.bw = a;
            this.g("ThumbCalloutFontWeight");
        }
        get resolvingAxisValueRef() {
            return this.bt;
        }
        set resolvingAxisValueRef(a) {
            this.bt = a;
            this.g("ResolvingAxisValueRef");
        }
        get windowRectChangedRef() {
            return this.by;
        }
        set windowRectChangedRef(a) {
            this.by = a;
            this.g("WindowRectChangedRef");
        }
    }
    ZoomSliderDescription.$t = /*@__PURE__*/ markType(ZoomSliderDescription, 'ZoomSliderDescription', Description.$);
    return ZoomSliderDescription;
})();
